// syracuse-ui/lib/tests/boost/boostProcessTest.js
"use strict";
var helpers = require("syracuse-core/lib/helpers");
var TestPage = require("syracuse-ui/lib/tests/testPage").TestPage;

var _processes = {
	VisualProcessACH05_002: "ACH05_002",
	VisualProcessActions: "Actions",
	VisualProcessLayers: "Layers",
	VisualProcessLines: "Lines",
	VisualProcessLineCapsJoints: "Linecaps & Linejoints",
	VisualProcessArrows: "Arrows",
	VisualProcessBezier: "Bezier",
	VisualProcessApiDashed: "Apishapes Dashed Borders",
	VisualProcessBzDashed: "Beziershapes Dashed Borders",
	VisualProcessBezierFill: "Fill Bezier ",
	VisualProcessPattern: "Fill with Pattern ",
	VisualProcessPatternRotation: "Pattern+ Rotation of shapes",
	VisualProcessRadialGradient: "Radial Gradient ",
	VisualProcessLinearGradientTransition: "LinearGradient Transition Vector  ",
	VisualProcessLinearGradientStops: "LinearGradient Stop Colors ",
	VisualProcessLinearGradientRotation: "LinearGradient + Rotation of shapes",
	VisualProcessTextRotation: "Text + Rotation of Shapes  ",
	VisualProcessFilling: "Filling",
	VisualProcessTextPositionning: "Text Positionning ",
	VisualProcessShadow: "Shadow ",
	VisualProcessTextFormatting: "Text Formatting ",
	VisualProcessTextSplitting: "Text Splitting",
	VisualProcessApiRotations: "Apishape Rotations",
	VisualProcessImgRotations: "Image Rotations",
	VisualProcessImgScaling: " Image scaling",
	VisualProcessImgRotationsDeformations: "image rotations and deformations",
	VisualProcessLinks: "Links",
	VisualProcessLinks2: "Links2",
	VisualDemoProcess: "Visual Process Demo",
	VisualDemoProcess2: "Visual Process Demo2",
	VisualProcess: "Test VP "
};

function ProcessTest() {}

exports.ProcessTest = helpers.defineClass(ProcessTest, TestPage, {
	onProcessClick: function(event, target) {
		var self = this;
		require.async("./process/" + target.syraValue, function(err, module) {
			var $resource = helpers.object.clone(module.$resources, true);
			syra_site.dom.empty(self.flashImage);
			self.flashImage.style.display = "none";
			self.flashImage.style.overflow = "visible";
			self.flashImage.style.padding = "0px";
			self.applyChange({
				$: {
					description: {
						$isHidden: !$resource.description
					}
				}
			});
			self.applyChange($resource);
			if ($resource.flashImage) {
				syra_site.dom.empty(self.flashImage);
				self.flashImage.appendChild(document.createElement("img")).setAttribute("src", $resource.flashImage.$url);
				self.flashImage.style.display = "";
			}
		});
	},
	loadBox: function() {
		this.$prototype = this.$prototype || {};
		this.$prototype.$properties = this.$prototype.$properties || {};
		this.$prototype.$properties.description = {
			$title: "Case description",
			$type: "application/x-string"
		};
		this.$prototype.$properties.process = {
			$type: "application/x-process",
			$title: "Visual Process Player test"
		};
		this.$item.$layout = {
			$items: {}
		};
		TestPage.prototype.loadBox.call(this);
		var keys = Object.keys(_processes);
		this.btns = {};
		for (var ii = 0, jj = keys.length; ii < jj; ii++) {
			var key = keys[ii];
			var btn = syra_menus.addTextButton(_processes[key], "s-mn-extra-link", "onProcessClick");
			this.layoutContent.layoutSlot.appendChild(this.btns[btn.syraValue = key] = btn);
			btn.style.display = "inline-block";
			btn.style.margin = "4px";
		}

		this.loadNewItem(this.layoutContent.layoutSlot, {
			$bind: "description",
			$isTitleHidden: true,
			$isEditMode: false
		});
		this.flashImage = this.layoutContent.layoutSlot.appendChild(document.createElement("div"));
		this.loadNewItem(this.layoutContent.layoutSlot, {
			$bind: "process",
			$isTitleHidden: true,
			$isEditMode: false
		});

	}
});
