// syracuse-ui/lib/tests/boost/boostProcessTest.js
"use strict";
var helpers = require("syracuse-core/lib/helpers") ;
var Article = require("syracuse-ui/lib/article/article").Article;
var util = require('syracuse-ui/lib/fusion/tools/util');



var _processes = {
    "testVisualProcessActions": {
        "$title": "Actions",
        "$resource": require("./process/VisualProcessActions").$resources
    },
    "testVisualProcessLayers": {
        "$title": "Layers",
        "$resource": require("./process/VisualProcessLayers").$resources
    },
    "testVisualProcessLines": {
        "$title": "Lines",
        "$resource": require("./process/VisualProcessLines").$resources
    },
    "testVisualProcessLineCapsJoints": {
        "$title": "Linecaps & Linejoints",
        "$resource": require("./process/VisualProcessLineCapsJoints").$resources
    },
    "testVisualProcessArrows": {
        "$title": "Arrows",
        "$resource": require("./process/VisualProcessArrows").$resources
    },
    "testVisualProcessBezier": {
        "$title": "Bezier",
        "$resource": require("./process/VisualProcessBezier").$resources
    },
    "testVisualProcessApiDashed": {
        "$title": "Apishapes Dashed Borders",
        "$resource": require("./process/VisualProcessApiDashed").$resources
    },
    "testVisualProcessBzDashed": {
        "$title": "Beziershapes Dashed Borders",
        "$resource": require("./process/VisualProcessBzDashed").$resources
    },
    "testVisualProcessBezierFill": {
        "$title": "Fill Bezier ",
        "$resource": require("./process/VisualProcessBezierFill").$resources
    },
    "testVisualProcessPattern": {
        "$title": "Fill with Pattern ",
        "$resource": require("./process/VisualProcessPattern").$resources
    },
    "testVisualProcessPatternRotation": {
        "$title": "Pattern+ Rotation of shapes",
        "$resource": require("./process/VisualProcessPatternRotation").$resources
    },
    "testVisualProcessRadialGradient": {
        "$title": "Radial Gradient ",
        "$resource": require("./process/VisualProcessRadialGradient").$resources
    },
    "testVisualProcessLinearGradientTransition": {
        "$title": "LinearGradient Transition Vector  ",
        "$resource": require("./process/VisualProcessLinearGradientTransition").$resources
    },
    "testVisualProcessLinearGradientStops": {
        "$title": "LinearGradient Stop Colors ",
        "$resource": require("./process/VisualProcessLinearGradientStops").$resources
    },
    "testVisualProcessLinearGradientRotation": {
        "$title": "LinearGradient + Rotation of shapes",
        "$resource": require("./process/VisualProcessLinearGradientRotation").$resources
    },
    "testVisualProcessTextRotation": {
        "$title": "Text + Rotation of Shapes  ",
        "$resource": require("./process/VisualProcessTextRotation").$resources
    },
    "testVisualProcessFilling": {
        "$title": "Filling",
        "$resource": require("./process/VisualProcessFilling").$resources
    },
    "testVisualProcessTextPositionning": {
        "$title": "Text Positionning ",
        "$resource": require("./process/VisualProcessTextPositionning").$resources
    },
    "testVisualProcessShadow": {
        "$title": "Shadow ",
        "$resource": require("./process/VisualProcessShadow").$resources
    },
    "testVisualProcessTextFormatting": {
        "$title": "Text Formatting ",
        "$resource": require("./process/VisualProcessTextFormatting").$resources
    },
    "testVisualProcessApiRotations": {
        "$title": "Apishape Rotations",
        "$resource": require("./process/VisualProcessApiRotations").$resources
    },
    "testVisualProcessImgRotations": {
        "$title": "Image Rotations",
        "$resource": require("./process/VisualProcessImgRotations").$resources
    },
    "testVisualProcessImgScaling": {
        "$title": " Image scaling",
        "$resource": require("./process/VisualProcessImgScaling").$resources
    },
    "testVisualProcessImgRotationsDeformations": {
        "$title": "image rotations and deformations",
        "$resource": require("./process/VisualProcessImgRotationsDeformations").$resources
    },
    "testVisualProcessLinks": {
        "$title": "Links",
        "$resource": require("./process/VisualProcessLinks").$resources
    },
    "testVisualProcessShapePositionning": {
        "$title": "Shape Positionning",
        "$resource": require("./process/VisualProcessShapePositionning").$resources
    },
    "demoVisualProcess": {
        "$title": "Visual Process Demo",
        "$resource": require("./process/VisualDemoProcess").$resources
    },
    "testVisualProcess": {
        "$title": "Test VP ",
        "$resource": require("./process/VisualProcess").$resources
    }
};


function ProcessTest(){
}

exports.ProcessTest = helpers.defineClass(ProcessTest, Article, {
    getTypes: function(){
        return Object.keys(_processes).map(function($type){
            return {
                $value: $type,
                $title: _processes[$type].$title
            };
        });
    },
    loadBox: function(){
        var self = this;
        self.$layoutOptions = document.site.$item.$layoutOptions.page["vignette"];
        self.$prototype = {
            $properties: {
                $field$type: {
                    $title: "select a Process type",
                    $isMetaData: true,
                    $type: "application/x-choice",
                    $value: {
                        $type: "application/x-string",
                        $enum: this.getTypes()
                    }
                },
                description: {
                    $title: "Case description",
                    $type: "application/x-string"
                },
                process: {
                    $type: "application/x-process",
                    $title: "Visual Process Player test"
                }
            
            }
        };
        self.$item = {
            $layout: {
                $items: [{
                    $bind: "$field$type",
                    $isEditMode: true,
                    $format: "$radios"
                }, {
                    $bind: "description",
                    $isTitleHidden: true,
                    $isEditMode: false
                }, {
                    $id: "flashImage",
                    $isTitleHidden: true,
                    $layout: {
                        $items: []
                    }
                }, {
                    $bind: "process",
                    $isTitleHidden: true,
                    $isEditMode: false,
                    $labelWidth: "auto"
                }]
            },
            $field$type: "testVisualProcessArrows"
        };
        Article.prototype.loadBox.call(self);
        self.notifyChange("testVisualProcessArrows", "$field$type");
    },
    notifyChange: function($bind, value){
        switch ($bind) {
            case "$field$type":
                //var $resource = _processes[value].$resource;
                var $resource = util.duplicateObj(_processes[value].$resource);

                var $$flasImage = this.idMap["flashImage"].$$body;
                $$flasImage.empty().hide();
                $$flasImage[0].style.overflow = "visible";
                $$flasImage[0].style.padding = "0px";
                this.applyChange({
                    $: {
                        description: {
                            $isHidden: !$resource.description
                        }
                    }
                });
                this.applyChange($resource);
                if ($resource.flashImage) {
                    $("<img />").appendTo($$flasImage.empty()).attr("src", $resource.flashImage.$url);
                    $$flasImage.show();
                }
                break;
        }
    }
});
