// syracuse-ui/lib/site/filter/filterMaker.js
"use strict";
var Parser = require('syracuse-sdata/lib/parser/parser').Parser;

var _operators = {
	like: {
		unFormat: function(value) {
			if (value && value.length >= 2) {
				if (value[0] == "%") {
					value = value.slice(1);
				}
				if (value[value.length - 1] == "%") {
					value = value.slice(0, value.length - 1);
				}
			}
			return value;
		},
		getFieldValue: function(options) {
			options.getRawValue = true;
			options.value = "'%" + ("" + _formatFieldValue(options)).replace(/'/g, "''") + "%'";
			return true;
		}
	},
	like_s: {
		unFormat: function(value) {
			if (value && value.length >= 1) {
				value = value.slice(0, value.length - 1);
			}
			return value;
		},
		getFieldValue: function(options) {
			options.getRawValue = true;
			options.value = "'" + ("" + _formatFieldValue(options)).replace(/'/g, "''") + "%'";
			options.operator = "like";
			return true;
		}
	},
	between: {
		getFieldValue: function(options) {
			var highValue = options.field.articleParent.dataset[options.$bind + "$High"];
			if (value && !highValue) {
				options.value = _formatFieldValue(options);
				options.operator = "ge";
			} else {
				if (!value && highValue) {
					options.value = highValue;
					options.value = _formatFieldValue(options);
					operator = "le";
				} else {
					if (value && highValue) {
						var minValue = _formatFieldValue(options);
						options.value = highValue;
						options.value = minValue + " and " + _formatFieldValue(options);
					} else {
						return false;
					}
				}
			}
			return true;
		}
	},
	empty: {
		getFieldValue: function(options) {
			options.value = "''";
			options.operator = "eq";
			return true;
		}
	},
	notempty: {
		getFieldValue: function(options) {
			options.value = "''";
			options.operator = "ne";
			return true;
		}
	}
};

function _formatFieldValue(options) {
	switch (options.$type) {
		case "application/x-reference":
			return _formatFieldValue({
				field: options.field,
				$bind: options.$bind,
				$type: options.$filterField.$type,
				value: options.field.getInputValue(),
				getRawValue: options.getRawValue
			});
		case "application/x-choice":
			var $field = options.field.$field;
			return _formatFieldValue({
				field: options.field,
				$bind: options.$bind,
				$type: ($field && $field.$value && $field.$value.$type) || "application/x-string",
				value: options.field.articleParent.dataset[options.$bind],
				getRawValue: options.getRawValue
			});
		case "application/x-boolean":
		case "application/x-integer":
		case "application/x-decimal":
		case "application/x-quantity":
			return options.field.articleParent.dataset[options.$bind];
		case "application/x-date":
			options.value = options.field.articleParent.dataset[options.$bind];
			return options.getRawValue ? options.value : ("@" + options.value + "@");
	}
	return options.getRawValue ? options.value : ("'" + (options.value || "") + "'");
}

exports.unFormatValue = function(operator, value) {
	var op = _operators[operator];
	return op && op.unFormat ? op.unFormat(value) : value;
};

function _findExpression(exp, $bind) {
	var found;
	for (var ii = 0, jj = exp.children.length; !found && ii < jj; ii++) {
		var child = exp.children[ii];
		if (child.type == "identifier" && child.value == $bind) {
			found = exp;
		} else {
			if (child.children) {
				found = _findExpression(child, $bind);
			}
		}
	}
	return found;
}


function _findExpressionOld(exp, $bind) {
	var found;
	for (var ii = 0, jj = exp.children.length; !found && ii < jj; ii++) {
		var child = exp.children[ii];
		if (child.children) {
			if (child.children[0].type == "identifier" && child.children[0].value == $bind) {
				found = child;
			} else {
				found = _findExpression(child, $bind);
			}
		}
	}
	return found;
}

function _ensureFilterCode(options) {
	if (options.value == "''" && options.$filterCode.indexOf(".") != -1) {
		options.value = null;
		options.$filterCode = options.field.$item.$bind;
	}
}

exports.parseWhere = function(where, $bind) {
	var exp = Parser.parse(where);
	if (exp && $bind) {
		exp = _findExpression(exp, $bind);
	}
	return exp;
};


exports.getFieldValue = function(options) {
	if (options.field.page.isFusionPage) {
		return syra_site.fusionGateway.filterMaker.getFieldValue(options);
	}
	options.value = options.field.getInputValue();
	if (options.value != null) {
		var op = _operators[options.operator];
		if (op && op.getFieldValue) {
			if (!op.getFieldValue(options)) {
				return "";
			}
		} else {
			options.value = _formatFieldValue(options);
		}

		// handling specific reference field case (fix SAM 97863)
		if (options.field && options.field.$field.$type == "application/x-reference") {
			_ensureFilterCode(options);
		}
		return "(" + options.$filterCode + " " + options.operator + " " + options.value + ")";
	}
};

exports.getFieldOperators = function($type, field, $filterField) {
	if (field.page.isFusionPage) {
		return syra_site.fusionGateway.filterMaker.getFieldOperators($type, field, $filterField);
	}
	switch ($type || field.$field.$type) {
		case "application/x-reference":
			return exports.getFieldOperators($filterField.$type, field);
		case "application/x-date":
		case "application/x-time":
		case "application/x-datetime":
			var operators = ["none", "gt", "ge", "lt", "le", "eq", "ne", "empty", "notempty"];
			if (!field.$field.$type == "application/x-reference") {
				operators.push("between");
			}
			return operators;
		case "application/x-string":
			var operators = ["none", "gt", "ge", "lt", "le", "eq", "ne", "like_s", "like", "empty", "notempty"];
			if (!field.$field.$type == "application/x-reference") {
				operators.push("between");
			}
			return operators;
		case "application/x-decimal":
		case "application/x-integer":
		case "application/x-quantity":
			if (!field.$field.$type == "application/x-reference") {
				return ["none", "eq", "ne", "empty", "notempty"];
			} else {
				return ["none", "gt", "ge", "lt", "le", "eq", "ne", "empty", "notempty"];
			}
			break;
	}
	return ["none", "eq", "ne", "empty", "notempty"];
};
