// msoffice/lib/ppt/ui/save.js
"use strict";

window.getStack = function() {
	try {
		var stack = [];
		var f = arguments.callee.caller;
		while (f) {
			var fn = f.toString();
			var fname = fn.substring(fn.indexOf('function') + 8, fn.indexOf('(')) || 'anonymous';
			stack.push(fname);
			f = f.caller;
		}
		return stack.join('\n');
	} catch (e) {}
};

window.onerror = function(message, url, line) {
	alert(message + '\nurl=' + url + '\nat line ' + line + '\n' + window.getStack());
	return false;
};

var siteFactory = require('syracuse-ui/lib/office/ppt/pptSaveSite');
exports.main = function() {
	siteFactory.load({
		$iconPath: "/syracuse-ui/themes/desktop/sage/images/",
		$pageTemplateUrl: "/sdata/syracuse/collaboration/syracuse/pages('{endpoint}.{representation},{target},{pageview}')?role={$role}",
		$userProfileUrl: "/sdata/syracuse/collaboration/syracuse/userProfiles/$template/$workingCopies?representation=userProfile.$edit&role={$role}",
		$syracuseMainPageUrl: "/syracuse-main/html/main.html"
	}, {
		$links: {
			$home: {
				$title: "Home",
				$url: "?representation=ppthome.$dashboard"
			}
		},
		$properties: {}
	});
};

if (require.main == module) exports.main();
