// syracuse-ui/lib/site/search/vignette/searchVignettePageResult.js
"use strict";
var helpers = require("syracuse-core/lib/helpers");
var DesktopPage = require('syracuse-ui/lib/page/desktopPage').DesktopPage;

function SearchVignettePageResult(){
}

exports.SearchVignettePageResult = helpers.defineClass(SearchVignettePageResult, DesktopPage, {
    applyChange: function(parent, data){
        DesktopPage.prototype.applyChange.call(parent, data);
        parent._displaySearchFacetsBar();
		parent.onWindowResize();
    },
    _displaySearchFacetsBar: function(){
        this.menuBar.isSlotVisible = true;
        if (!this.menuBar.page.designer) {
            this.menuBar.barSlot.style.display = "";
            if (this.menuBar.page.layoutSlot.style.display != "none") {
                document.site.resize();
            }
        }
    },
    _ensureMenus: function(){
        DesktopPage.prototype._ensureMenus.call(this);
        if (this.menuBar._bar) {
            this.loadNewItem(this.menuBar._bar, {
                $category: "searchFacets",
                $prototype: {},
                $bind: "$searchFacets"
            });
        }
    },
    dispose: function(){
        DesktopPage.prototype.dispose.call(this);
    }
});
