// syracuse-ui/lib/site/search/vignette/searchVignettePageBrowse.js
"use strict";
var helpers = require("syracuse-core/lib/helpers");
var DesktopPage = require('syracuse-ui/lib/page/desktopPage').DesktopPage;

function SearchVignettePageBrowse(){
}

exports.SearchVignettePageBrowse = helpers.defineClass(SearchVignettePageBrowse, DesktopPage, {
    loadBox: function(vignetteData, articleParent){
        this.$skin = "s-searchvignette-browse";
        this.articleParent = articleParent;
        this.vignetteData = vignetteData;
        var initData = this.buildInitData(vignetteData);
        this._defaultTitle = "Search vignette";
        
        // TODO
        //this.$prototype = this.getPrototype();
        this.$item = this.getItem();
        
        this.layoutSlot = document.createElement("div");
        this.layoutSlot.className = this.$skin;
        
        // append to search vignette layout
        articleParent.scrollview.appendChild(this.layoutSlot);
        
        DesktopPage.prototype.loadBox.call(this);
    },
    buildInitData: function(vignetteData){
        var moduleKey = Object.keys(vignetteData.modules);
        var modules = [];
        for (var ii = 0, jj = moduleKey.length; ii < jj; ii++) {
            modules.push({
                title: vignetteData.modules[moduleKey[ii]].title
            });
        }
        return {
            modules: modules
        };
    },
    getItem: function(){
        return {
            $category: "section",
            $isTitleHidden: true,
            $layout: {
                $items: [{
                    $layoutType: "stack"
                }]
            }
        };
    }
});
