// syracuse-ui/lib/page/navigation/subModule/subModuleList.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var DesktopList = require("syracuse-ui/lib/field/array/list").DesktopList;
var SubModuleBuilder = require("./subModuleBuilder").SubModuleBuilder;

function SubModuleList() {}

exports.SubModuleList = helpers.defineClass(SubModuleList, DesktopList, {
	loadBox: function(initData) {
		this.isSubModuleList = true;
		this.emptyMessage = this.localize.dskpNavigationPageNoSubModule;
		this.$confirmDelete = this.localize.dskpNavigationPageConfirmDeleteSubModule;
		this.BuilderClass = SubModuleBuilder;
		this.$isQuickDesignerEnabled = false;
		this.$item.$isPagerHidden = true;
		DesktopList.prototype.loadBox.call(this, initData);
		this.topbar.setAttribute("data-s-inout", this.id);
		this._menusSlot.style.visibility = "hidden";
	},
	onItemEnterLeave: function(event) {
		if (this._menusSlot) {
			this._menusSlot.style.visibility = event.type == "mouseenter" ? "visible" : "hidden";
		}
	},
	_createMenusBox: function() {
		if (!this.isMenuLoaded) {
			this.isMenuLoaded = true;
			this.page.loadNewItem(this._menusSlot, {
				$layoutType: "row",
				$autoSize: true,
				$items: [{
					$skin: "s-list-menus-link",
					$bind: "$create",
					$category: "link",
					$icon: {
						$mode: "icon",
						$path: "list/s-list-"
					}
				}, {
					$skin: "s-list-menus-link",
					$bind: "$select",
					$category: "link",
					$icon: {
						$mode: "icon",
						$path: "list/s-list-"
					}
				}]
			}, this);
		}
	},
	dispose: function() {
		document.site.emptyDom(this._core);
		DesktopList.prototype.dispose.call(this);
	}
});
