// syracuse-ui/lib/page/navigation/subModule/subModuleList.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var DesktopList = require("syracuse-ui/lib/field/array/list").DesktopList;
var SubModuleBuilder = require("./subModuleBuilder").SubModuleBuilder;

function SubModuleList() {}

exports.SubModuleList = helpers.defineClass(SubModuleList, DesktopList, {
	loadBox: function(initData) {
		this.BuilderClass = SubModuleBuilder;
		this.$isQuickDesignerEnabled = false;
		this.$item.$isPagerHidden = true;
		DesktopList.prototype.loadBox.call(this, initData);
	},
	_createMenusBox: function() {
		var $menuBag = {
			$category: "menus",
			$id: this.id + "-menu-picker",
			$skin: "s-list-menus",
			$itemSkin: "s-record-menus-link",
			$layout: {
				$layoutType: "row",
				$autoSize: true,
				$items: [{
					$skin: "s-list-menus-link",
					$bind: "$create",
					$category: "link",
					$icon: {
						$mode: "icon",
						$path: "list/s-list-"
					}
				}, {
					$skin: "s-list-menus-link",
					$bind: "$select",
					$category: "link",
					$icon: {
						$mode: "icon",
						$path: "list/s-list-"
					}
				}]
			}
		};
		this.page.loadNewItem(this._menusSlot, $menuBag, this);
		this.menusBox = this.idMap[this.id + "-menu-picker"];
		this.menusBox.contextField = this;
	},
	dispose: function() {
		document.site.emptyDom(this._core);
		DesktopList.prototype.dispose.call(this);
	}
});
