// @sage/syracuse-react/src/common/list/list.tsx
"use strict";
import * as React from 'react';
import Scrollbar from '../scroll/scrollbar/scrollbar';
import * as ReactDom from 'react-dom';

export class ScrollList extends React.Component<{
    className: string;
    selectedIndex?: number;
}, {}>{
    private _rootNode: HTMLElement;
    get rootNode(): HTMLElement {
        if (!this._rootNode) {
            this._rootNode = ReactDom.findDOMNode(this) as HTMLElement;
        }
        return this._rootNode;
    }
    set listHeight(maxHeight: number) {
        (this.rootNode.firstChild as HTMLElement).style.maxHeight = maxHeight + "px";
    }
    componentWillUnmount() {
        this._rootNode = null;
    }
    render() {
        return <Scrollbar>
            <ul className={this.props.className}>
                {React.Children.map(this.props.children, (child: React.ReactChild, index: number) => {
                    let css = this.props.className + "_li";
                    if (this.props.selectedIndex == index) {
                        css += " s_selected";
                    }
                    return <li key={index} className={css} >{child}</li>;
                })}
            </ul>
        </Scrollbar>;
    }
}
