// syracuse-ui/lib/tests/pages/trackers/trackers.js
"use strict";
var _tests = [require("./searchIndex")];
var _sample = require("./sample");

exports.load = function(page, finish) {
	page.$prototype.$properties.mode = {
		$title: "mode",
		$type: "application/x-choice",
		$value: {
			$type: "application/x-string",
			$enum: [{
				$value: "auto",
				$title: "auto"
			}, {
				$value: "manual",
				$title: "manual"
			}]
		}
	};

	page.loadNewItem(page.leftCol, {
		$isEditMode: true,
		$isTopLabelAlignment: false,
		$bind: "mode"
	});
	page.applyChange({
		mode: "auto"
	});

	function _onTrackerClick() {
		var btn = this;
		var test = _tests[btn.textIndex];
		if (!btn.step) {
			btn.interval = btn.step = 0;
			var data = test.trackers[btn.step];
			data.isTestCase = true;
			test.tracker = syra_trackers.addTracker(data);
		}
		if (btn.parent.boundFields.mode[0].getValue() == "auto") {
			btn.step++;
			btn.interval = setInterval(function() {
				test.tracker.applyChange(test.trackers[btn.step++], btn.step == test.trackers.length);
				if (btn.step == test.trackers.length) {
					clearInterval(btn.interval);
					test.tracker = btn.step = btn.interval = 0;
				}
			}, 800);
		} else {
			syra_button.setText(btn, "Stop " + test.$title + " ( " + (btn.step + 1) + "/" + test.trackers.length + ")");
			if (btn.step == 0) {
				btn.step++;
				return;
			}
			test.tracker.applyChange(test.trackers[btn.step++], btn.step == test.trackers.length);
			if (btn.step == test.trackers.length) {
				test.tracker = btn.step = 0;
				syra_button.setText(btn, test.$title + " ( " + test.trackers.length + ")");
			}
		}
	};

	for (var ii = 0, jj = _tests.length; ii < jj; ii++) {
		syra_button.add({
			parent: page,
			slot: page.leftCol,
			text: _tests[ii].$title + " ( " + _tests[ii].trackers.length + ")",
			css: "s-uitest-mn",
			btnclick: _onTrackerClick,
			textIndex: ii
		});
	}

	syra_button.add({
		parent: page,
		slot: page.rightCol,
		text: _sample.$title,
		css: "s-uitest-mn",
		btnclick: function() {
			var step = 0;
			var interval = setInterval(function() {
				var data = _sample.trackers[step++];
				data.isTestCase = true;
				syra_trackers.addTracker(data);
				if (step == _sample.trackers.length) {
					clearInterval(interval);
					step = interval = 0;
				}
			}, 800);
		}
	});

	page.setTitle("Trackers");
	page.notifyDataChange = function(field, value) {

	};

	finish(page);
};
