// syracuse-tablet/html/js/demos/clientTestApp/unitTests/entities/entityNativeFields.js
"use strict";

var _entityName = "nativeFields";
var _unitTestCallBack = function(structElmt) {

};

var _entity = {
	$entityName: _entityName,
	$title: "Test native fields",
	$properties: {
		"barcode": {
			"$title": "Barcode",
			"$type": "application/x-string"
		},
		"imageOriginal": {
			"$title": "Image Original",
			"$type": "image"
		},
		"imageSmall": {
			"$title": "Image Small",
			"$type": "image"
		},
		"imageMedium": {
			"$title": "Image Medium",
			"$type": "image"
		},
		"imageLarge": {
			"$title": "Image Large",
			"$type": "image"
		}
	},
	$relations: {},
	$articles: {
		$edit: {
			"$title": "Test barcode",
			"$layoutType": "stack",
			"$items": [{
				"$bind": "barcode",
				"$scanBarcode": true
			}, {
				"$bind": "imageOriginal",
				"$imgScale": "original"
			}, {
				"$bind": "imageSmall",
				"$imgScale": "small"
			}, {
				"$bind": "imageMedium",
				"$imgScale": "medium"
			}, {
				"$bind": "imageLarge",
				"$imgScale": "large"
			}]
		}
	},
	$services: {
		$read: function(op, data) {
			var deferred = $.Deferred();
			var result = null;
			deferred.resolve($.extend({}, result));
			return deferred.promise();
		}
	},
	//sdataUrl 	parsed url
	//path 		array that contains th end of the path following entity name
	$urlCall: function(sdataUrl, path) {
		return "";
	}
};


exports.entity = _entity;
