// syracuse-ui/lib/desktop/field/booleanField.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var commonField = require('syracuse-ui/lib/common/field/booleanField');

function DesktopBoolean(){
}

exports.DesktopBoolean = helpers.defineClass(DesktopBoolean, commonField.BooleanField, {
    applyDesignMetaData: function(metadata, onAuthoring){
        commonField.BooleanField.prototype.applyDesignMetaData.call(this, metadata, onAuthoring);
        if (metadata.$isLeftSpaceHidden !== undefined) {
            this.$$titleSpace[0].style.display = metadata.$isLeftSpaceHidden ? "none" : "";
        }
    },
    render: function(){
        if (this.$item.$inplace) {
            this.$$item.css("text-align", "center");
        }
        if (this.$isEditMode) {
            this.$$input = $("<input type='checkbox'/>").addClass(this.$skinInput + "-boolean").appendTo(this.$$value);
            if (this.$$title) {
                var title = this.$$title[0];
                var div = document.createElement("div");
                div.className = title.className;
                this.$$titleSpace = $(div).prependTo(this.$$item);
                var css = " " + this.$skin + "-title-boolean";
                if (this.$item.$css) {
                    css += " " + this.$item.$css;
                }
                title.className = css;
                this.$$value.append(this.$$title);
            }
        }
        else {
            this.$$value.addClass("s-boolean");
        }
    },
    hideTitle: function($isTitleHidden){
    
    }
});
