// syracuse-ui/lib/desktop/field/datetimeField.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var picker = require('./datetimePicker');
var DateTimeField = require('syracuse-ui/lib/common/field/datetimeField').DateTimeField;

function DesktopDateTime(){
}

exports.DesktopDateTime = helpers.defineClass(DesktopDateTime, DateTimeField, {
    clickPicker: function($$btn){
        var self = this;
        if (!self.$isDisabled) {
            if (!self.popupPicker) {
                self.$$item.attr("id", self.id);
                self.focus();
                self.picker = new picker[$$btn[0].className.indexOf("-time-picker") > 0 ? "TimeChoice" : "Calendar"]();
                self.picker.create(self);
                self.popupPicker = document.site.openDialog({
                    $dialogMode: "popup",
					content: self,
                    $$popup: self.picker.$$item,
                    position: {
                        my: "right top",
                        at: "right bottom",
                        of: self.$$input
                    },
                    closeHandler: function(){
                        setTimeout(function(){
                            if (self.picker) {
                                self.picker.remove();
                                self.picker = null;
                            }
                            self.popupPicker = null;
                        }, 200);
                        return true;
                    },
                    boundarySelector: "#" + self.id
                });
            }
            else {
                self.popupPicker.close();
            }
        }
    },
    setPickerValue: function(value){
        this.popupPicker.close();
        this.$$input.val(value).focus().change();
    },
    render: function(){
        var self = this;
        if (self.$isEditMode) {
            // application/x-{date,time}
            if (self.$field.$type != "application/x-datetime") {
                var cssSuffix = (self.$field.$type == "application/x-time" ? "-time-picker" : "-date-picker");
                var input = (self.$$input = $("<input type='text'/>"))[0];
                input.className = self.$skinInput + " " + self.$skinInput + cssSuffix;
                var btn = document.createElement("a");
                btn.className = self.$skin + cssSuffix;
                self.$$pickerBtn = $(btn).bind("click", function(){
                    self.clickPicker($(this));
                });
                var box = document.createElement("div");
                box.className = self.$skinInput + "-box";
                $(box).append(self.$$input).append(self.$$pickerBtn).appendTo(self.$$dataValue);
            }
            // application/x-datetime
            else {
                var input = (self.$$input = $("<input type='text'/>"))[0];
                input.className = self.$skinInput
                var box = document.createElement("div");
                box.className = self.$skinInput + "-box";
                $(box).append(self.$$input).appendTo(self.$$dataValue);
            }
        }
        else {
            DateTimeField.prototype.render.call(this);
        }
    }
});
