// syracuse-development/lib/representations/quality/package.js
"use strict";

exports.representation = {
	$entityName: "package",
	$facets: {
		$details: {
			$layout: {
				"$layoutType": "stack",
				"$items": [{
					"$category": "section",
					"$title": "Information",
					"$layout": {
						"$items": [{
							"$bind": "name"
						}, {
							"$bind": "description"
						}],
						"$layoutType": "stack"
					}
				}, {
					"$category": "section",
					"$title": "Unit Tests",
					"$layout": {
						"$items": [{
							"$bind": "unitTests",
							"$isQuickFilter": true,
							"$layout": {
								"$items": [{
									"$bind": "name"
								}, {
									"$bind": "description"
								}, {
									"$bind": "active"
								}, {
									"$bind": "deleted"
								}]
							}
						}],
						"$layoutType": "stack"
					}
				}]
			}
		},
		$edit: {
			$copy: "$details"
		},
		$query: {
			$layout: {
				$items: [{
					$category: "section",
					$layout: {
						$items: [{
							$category: "section",
							$layout: {
								$items: [{
									$bind: "$resources",
									$layout: {
										$items: [{
											$bind: "name"
										}, {
											$bind: "description"
										}, {
											$bind: "active"
										}, {
											$bind: "deleted"
										}]
									}
								}]
							}
						}]
					}
				}]
			}
		}
	}
};
