// syracuse-development/lib/representations/quality/unitTest.js
"use strict";

exports.representation = {
	$entityName: "unitTest",
	$facets: {
		$details: {
			$layout: {
				"$layoutType": "stack",
				"$items": [{
					"$category": "section",
					"$title": "Information",
					"$layout": {
						"$items": [{
							"$bind": "package"
						}, {
							"$bind": "name"
						}, {
							"$bind": "description"
						}, {
							"$bind": "scope"
						}],
						"$layoutType": "stack"
					}
				}, {
					"$category": "section",
					"$title": "Status",
					"$layout": {
						"$items": [{
							"$bind": "active"
						}, {
							"$bind": "pass"
						}, {
							"$bind": "assertions"
						}, {
							"$bind": "errors"
						}, {
							"$bind": "lastRun"
						}, {
							"$bind": "results"
						}],
						"$layoutType": "stack"
					}
				}, {
					"$category": "section",
					"$title": "Paths",
					"$layout": {
						"$items": [{
							"$bind": "deleted"
						}, {
							"$bind": "module"
						}, {
							"$bind": "filename"
						}],
						"$layoutType": "stack"
					}
				}]
			}
		},
		$edit: {
			$copy: "$details"
		},
		$query: {
			$layout: {
				$items: [{
					$category: "section",
					$layout: {
						$items: [{
							$category: "section",
							$layout: {
								$items: [{
									$bind: "$resources",
									$layout: {
										$items: [{
											"$bind": "package"
										}, {
											"$bind": "name"
										}, {
											"$bind": "description"
										}, {
											"$bind": "scope"
										}, {
											"$bind": "active"
										}, {
											"$bind": "pass"
										}, {
											"$bind": "lastRun"
										}]
									}
								}]
							}
						}]
					}
				}]
			}
		}
	}
};
