// syracuse-tablet/html/js/controls/array/arrayBase.js
"use strict";


var utils = require('syracuse-tablet/html/js/helpers/utils');
var Base = require('syracuse-tablet/html/js/controls/ctrlBase').Klass;

/**
 * x-array control base class
 */
var _Klass = utils.defineClass(
	function(controller, article, prototype) {
		var self = this;
		Base.call(self, controller, article, prototype);
		self.builder = null;
	}, Base, {

		destroy: function() {
			var self = this;
			Base.prototype.destroy.call(self);
			self.prototype = null;
			if (self.builder) {
				self.builder.destroy();
				self.builder = null;
			}
			self.arrayData = null;
		},

		/**
		 *	Build html DOM object + bind event
		 *		arrayData is an object that contains all data needed by array
		 *		depends on the kind of page query or detail/edit
		 **/
		buildHtml: function($parent, arrayData, buildOptions) {
			var self = this;
			if (self.builder == null) throw new Error("Null builder");
			self.arrayData = arrayData;
			if (self.arrayData == null) throw new Error("Null arrayData");
			self.builder.buildHtml($parent, arrayData, buildOptions);
		},

		setBuilder: function(builder) {
			this.builder = builder;
		},

		/**
		 *	returns array rows
		 *		controllerDao optional
		 **/
		getArrayData: function() {
			return this.arrayData;
		}

	});

exports.Klass = _Klass;
