// syracuse-tablet/html/js/controls/array/arrayBase.js
"use strict";


var utils = require('syracuse-tablet/html/js/helpers/utils');
var Base = require('syracuse-tablet/html/js/controls/ctrlBase').Klass;

var _destroyArrayData = function(a) {
	if (!a || !a.$resources) return null;
	a.$resources.forEach(function(r) {
		if (r) r.destroy();
	});
	return null;
};
/**
 * x-array control base class
 */
var _Klass = utils.defineClass(
	function(controller, article, prototype) {
		var self = this;
		Base.call(self, controller, article, prototype);
		self.builder = null;
		self._arrayData = null;
	}, Base, {

		destroy: function() {
			var self = this;
			Base.prototype.destroy.call(self);
			self.prototype = null;
			if (self.builder) {
				self.builder.destroy();
				self.builder = null;
			}
			self._arrayData = _destroyArrayData(self._arrayData);
		},

		/**
		 *	Build html DOM object + bind event
		 **/
		buildHtml: function($$parent, controllerDao, buildOptions) {
			var self = this;
			if (self.builder == null) throw new Error("Null builder");
			self.builder.buildHtml($$parent, self._setArrayData(controllerDao), buildOptions);
		},

		setBuilder: function(builder) {
			this.builder = builder;
		},

		/**
		 * Refreshes the control
		 */
		refresh: function(controllerDao, options) {
			var self = this;
			if (self.builder == null) throw new Error("Null builder");
			self.builder.refresh(self._setArrayData(controllerDao), options);
		},

		/**
		 *	returns array rows
		 *		depends on the kind of page query/lookup or detail/edit
		 *		called one time to create the dao - then use getArrayData()
		 **/
		readArrayData: function(controllerDao) {
			throw new Error("not implemented");
		},

		/**
		 *	Set array data - depends on array type
		 **/
		_setArrayData: function(controllerDao) {
			var self = this;
			self.empty();
			_destroyArrayData(self._arrayData);
			self._arrayData = self.readArrayData(controllerDao);
			if (self._arrayData == null) throw new Error("Null arrayData");
			return self._arrayData;
		},

		/**
		 *	returns array rows
		 *		controllerDao optional
		 **/
		getArrayData: function() {
			return this._arrayData;
		}

	});

exports.Klass = _Klass;
