// syracuse-tablet/html/js/controls/array/builderHeader.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');

var _html = {
	header: '<header> \
			<nav class="s-m-stdheader"> \
				{{#if showPagination}} \
				<nav class="s-m-pagin"> \
					<ul class="pagination"> \
						<li> \
							<a class="s-m-link glyphicon glyphicon-step-backward disabled" href="#" \
								data-nevent \
								data-naction="pagination" \
								data-params="$first" \
								draggable="false"> \
							</a> \
						</li> \
						<li> \
							<a class="s-m-link glyphicon glyphicon-chevron-left disabled" href="#" \
								data-nevent \
								data-naction="pagination" \
								data-params="$previous" \
								draggable="false"> \
							</a> \
						</li> \
						<li> \
							<a class="s-m-link glyphicon glyphicon-chevron-right disabled" href="#" \
								data-nevent \
								data-naction="pagination" \
								data-params="$next" \
								draggable="false "> \
							</a> \
						</li> \
						<li> \
							<a class="s-m-link glyphicon glyphicon-step-forward disabled" href="#" \
								data-nevent \
								data-naction="pagination" \
								data-params="$last" \
								draggable="false"> \
							</a> \
						</li> \
					</ul> \
				</nav> \
				{{/if}} \
				{{#if showSearch}} \
					<nav class="s-m-search"/>\
				{{/if}} \
				{{#if showSortFilter}} \
					{{#if hasHeader}}\
						<div class="s-m-filterSort" data-nevent data-naction="actionArray" data-params="panelfiltersort">\
							<i class="s-m-sorted" style="display:none"></i>\
							<i class="s-m-filtered fa fa-filter" style="display:none"></i>\
							<i class="fa fa-bars"></i>\
						</div>\
					{{/if}}\
				{{/if}}\
				{{#if showFiltersList}}\
					<nav class="s-m-filterslist"/>\
				{{/if}}\
			</nav> \
			{{#if showFiltersTabs}}\
				<nav class="s-m-filtertabs"/>\
			{{/if}}\
		</header>'
};

var _templates = {};
var _getHtml = function(name, ctx) {
	var tmpl = _templates[name];
	if (!tmpl) {
		tmpl = _templates[name] = Handlebars.compile(_html[name]);
	}
	return tmpl(ctx || {});
};

/**
 * Render array header
 */
exports.BuilderHeader = utils.defineClass(
	function builderHeader(control, headerOpts) {
		this.control = control;
		this.options = headerOpts;
		this.options.showFiltersList = this.options.$filtersAuthoring === "list";
		this.options.showFiltersTabs = this.options.showFiltersList !== true && this.options.$filtersAuthoring === "tabs";
		if (this.options.showFiltersTabs == true && this.control.page.isVignette()) {
			this.options.showFiltersList = true;
			this.options.showFiltersTabs = false;
		}
		this.options.hasHeader = this.control.page.isVignette();
	}, null, {

		destroy: function() {
			this.control = null;
			if (this.$$elmt) {
				this.$$elmt.remove()
			}
			this.$$elmt = null;
		},

		buildHtml: function(arrayData, refresh) {
			// Hide search if no data and no search criteria
			var showSearch = (this.options.showSearch && (arrayData.length > 0 || this.control.filterSearch.hasSearchCriteria()));
			if (refresh == true && this.$$elmt) {
				this.update(showSearch);
				return;
			}
			if (this.$$elmt) {
				this.$$elmt.remove();
			}
			if (this.options.showFiltersList || this.options.showFiltersTabs || showSearch || this.options.showSortFilter || this.options.showPagination) {
				this.$$elmt = $(_getHtml("header", this.options));
				if (showSearch) {
					$("nav.s-m-search", this.$$elmt).append(this.control.filterSearch.getHtml())
				}
				if (this.options.showFiltersList) {
					$("nav.s-m-filterslist", this.$$elmt).append(this.control.filterX3.getHtml("list"));
				} else if (this.options.showFiltersTabs) {
					$("nav.s-m-filtertabs", this.$$elmt).append(this.control.filterX3.getHtml("tabs"));
				}
			} else {
				// Empty
				this.$$elmt = $("<header/>");
			}
			this.control.$$content.before(this.$$elmt);
			this.update(showSearch);
			// Check if header empty
		},

		update: function(showSearch) {
			$("nav.s-m-search", this.$$elmt).toggle(showSearch);
			if (this.control.filterSort) {
				this.control.filterSort.arrayHeaderUpdate(this.$$elmt);
			}
			if (this.control.paginator) {
				var self = this;
				var nbEnabled = 0;
				this.control.paginator.pagingLinks.forEach(function(link) {
					var $$link = $('[data-naction="pagination"][data-params="' + link + '"]', self.$$elmt);
					var ena = self.control.paginator.isEnabled(link);
					if (ena) {
						nbEnabled++;
					}
					$($$link.parent()).add($$link).toggleClass("disabled", !ena);
				});
				// All disabled hide pagination
				this.$$elmt.find(".s-m-pagin").toggle(nbEnabled > 0);
			}
		},
		/**
		 * disable/enable clicks on the panel
		 */
		disable: function(yes) {
			yes = yes === true;
			this.$$elmt.css("opacity", yes ? 0.5 : "inherit");
			if (yes) {
				this.$$elmt.click(function(evt) {
					evt.preventDefault();
					evt.stopPropagation();
				})
			} else {
				this.$$elmt.off();
			}
		}
	});
