// syracuse-tablet/html/js/controls/array/builderSeparator.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var uiUtils = require('syracuse-tablet/html/js/ui/uiUtils');
var Base = require('syracuse-tablet/html/js/controls/array/builderBase').BuilderBase;
var modules = require('syracuse-tablet/html/js/common/modules');
var SdataController = require("syracuse-tablet/html/js/controllers/sdataController").SdataController;

/**
 * Display the control with displayCtx = "table"
 * I didn't want to change the displayCtx in all fields control so I kept "table"
 */
var _createCellControl = function(arrayCtrl, $$parent, colInfo, fieldProto) {
	var article = {
		"$bind": colInfo.$bind
	};
	return ctrlFactory.createCellFieldCtrl($$parent, arrayCtrl.controller, colInfo.article || colInfo, fieldProto);
};
var _seps = {
	semicolon: ",",
	colon: ":",
	blank: " ",
	dash: "-"
};

exports.BuilderSeparator = utils.defineClass(
	function builderSeparator(control) {
		Base.call(this, control);
	}, Base, {

		destroy: function() {
			Base.prototype.destroy.call(this);
		},

		buildHtml: function(data, $$parent, refresh) {
			Base.prototype.buildHtml.call(this, data, $$parent, refresh);
			$$parent.empty();
			//this.control.arrayController.destroyChildren();
			var $$section = $('<section/>').appendTo($$parent);
			var sep = _seps[this.control.article.$separator || "blank"] || " ";
			var ctrlFactory = modules.get("ctrlFactory");
			var self = this;
			data.forEach(function(dataSet, idx) {
				if (idx > 0) {
					var $$sep = $('<span class="s-m-separator"/>').appendTo($$section).text(sep);
					if (sep === ",") {
						$$sep.css("text-align", "left");
					}
				}
				// We need to have one controller per field because the dataset is unic per field (same $bind)
				var controller = new SdataController(dataSet, self.control.arrayController);
				//self._fieldsControllers.push(controller);
				var ctrlFld = ctrlFactory.createControl(controller, {
						$bind: dataSet.parent.valueProperty
					},
					null,
					self.control.page, {
						displayCtx: "separator",
						noEdit: true
					});
				if (ctrlFld) {
					ctrlFld.set$$container($$section);
					ctrlFld.buildHtml();
				}
			});
		}

	});
