// syracuse-tablet/html/js/controls/array/ctrlArrayChart.js
"use strict";


var utils = require('syracuse-tablet/html/js/helpers/utils');
var Base = require('syracuse-tablet/html/js/controls/array/ctrlArray').Klass;
var globals = require('syracuse-tablet/html/js/helpers/globals');
var ctrlCubeChartInterface = require('syracuse-tablet/html/js/controls/chart/ctrlCubeChart').arrayChartInterface;

var _Klass = utils.defineClass(
	function CtrlArrayChart($type, controller, article, prototype, options) {
		Base.call(this, controller, article, prototype.create({
			$type: $type
		}), options);
	}, Base, {
		buildHtml: function($$parent, controllerDao, buildOptions) {
			Base.prototype.buildHtml.call(this, $$parent, controllerDao, buildOptions);
			this.$$elmt.attr("data-control-id", this.id);
			this.$$elmt.attr("data-action", globals.ACTCHARTDETAIL);
		},
		onResize: function(context) {
			Base.prototype.onResize.call(this, context);
			ctrlCubeChartInterface.notifChartHosted(this);
		},
		_actRowSelect: function(param, $value, event) {
			event.preventDefault();
			event.stopPropagation();
			this._actChartDetail();
		},
		_actChartDetail: function() {
			ctrlCubeChartInterface.openChartDetailPage(this);
		},
		childPageGetPageInfo: function(typePage) {
			return ctrlCubeChartInterface.childPageGetPageInfo(this, typePage);
		},
		setChartDetailAuthoring: function(article) {
			ctrlCubeChartInterface.setChartDetailAuthoring(this, article);
		}
	});

exports.Klass = _Klass;
