// syracuse-tablet/html/js/controls/array/filterHelper.js
"use strict";

var parser = require('syracuse-tablet/html/js/sdata/sdatawhere/parser').Parser;
var jsutils = require('syracuse-tablet/html/js/helpers/jsutils');

exports.getFilterWhereClause = function(filterInfo) {
	var filterClause = "";
	if (filterInfo != null && filterInfo.length != 0) {
		$.each(filterInfo, function(idx, filter) {
			var cf = '(' + filter.id + ' ' + (filter.operator === "like_s" ? "like" : filter.operator) + ' ' + _setFilterRightValue(filter) + ')';
			filterClause = filterClause.length == 0 ? cf : filterClause + ' and ' + cf;
		});
	}
	return filterClause;
}

exports.getSearchClause = function(searchInfo) {
	if (!searchInfo) {
		return null;
	}
	var searchClause = '';
	Object.keys(searchInfo).forEach(function(key) {
		var search = searchInfo[key];
		var cf = '(' + search.id + ' ' + search.operator + ' ' + _setFilterRightValue(search) + ')';
		searchClause = searchClause.length == 0 ? cf : searchClause + ' or ' + cf;
	});
	if (searchClause) {
		searchClause = '(' + searchClause + ')';
	}
	return searchClause;
}
exports.filterSortUpdateUrl = function(opt) {
	var options = opt || {};
	var sdataUrl = jsutils.parseURL(options.$url);
	if (options.$filters) {
		if (!sdataUrl.query) {
			sdataUrl.query = {};
		}
		sdataUrl.query.filter = options.$filters.id;
	}
	// Add filter/sort clause for Lookup and Query if any
	// build sort clause
	var sortClause = null;
	if (options.sortInfo != null && options.sortInfo.length != 0) {
		options.sortInfo.some(function(i) {
			if (i.sort != "none") {
				sortClause = {
					order: i.sort,
					field: i.id
				};
				return true;
			}
		});
	}
	var filterMsg = false;
	// build filter clause
	var filterClause = exports.getFilterWhereClause(options.filterInfos);
	var searchClause = exports.getSearchClause(options.searchInfo);
	if (searchClause && !exports.isValidWhere(searchClause)) {
		searchClause = null;
		filterMsg = true;
	}
	// apply sort clause
	if (!sortClause && !filterClause && !searchClause) return jsutils.urlToString(sdataUrl);;
	// sdataUrl.query.key != null means it's a pagination request
	// -> The pagination request contains already the sort criteria so we don't add them
	// -> Otherwise we add sort criteria to the request
	if (sortClause && sdataUrl.query.key == null) {
		var old = sdataUrl.query.orderBy;
		sdataUrl.query.orderBy = sortClause.field + ' ' + sortClause.order;
		if (old) sdataUrl.query.orderBy = ", " + old;
	}
	// apply filter clause
	if (filterClause && sdataUrl.query.key == null) {
		var old = sdataUrl.query.where;
		sdataUrl.query.where = filterClause;
		if (old) sdataUrl.query.where = old + ' and ' + sdataUrl.query.where;
	}

	// apply search clause
	if (searchClause && sdataUrl.query.key == null) {
		var old = sdataUrl.query.where;
		sdataUrl.query.where = searchClause;
		if (old) sdataUrl.query.where = old + ' and ' + sdataUrl.query.where;
	}

	return jsutils.urlToString(sdataUrl);
}
exports.isValidWhere = function(where) {
	try {
		parser.parse(where);
		return true;
	} catch (e) {
		return false;
	}
}

/**
 * 
 * Returns appropriate right value for where request
 * For example, for the filter 'CODE contains aa', the where clause should look like
 * CODE like '%aa%'. This function will return '%aa%'
 */
function _setFilterRightValue(filter) {
	var cp = filter.prop;
	var cv, value = (filter.originalValue || filter.value) == null ? "" : (filter.originalValue || filter.value);
	switch (cp.$type) {
		case "application/x-password":
		case "application/x-string":
			cv = value.replace(/'/g, "\\'");
			if (filter.operator === "like") cv = '%' + cv + '%';
			else if (filter.operator === "like_s") cv = cv + '%';
			cv = "'" + cv + "'";
			return cv;
		case "application/x-integer":
		case "application/x-real":
		case "application/x-decimal":
		case "application/x-quantity":
			return value;
		case "application/x-date":
			if (!value) return null;
			return '@' + value + '@';
		case "application/x-datetime":
			return '@' + value + '@';
		case "application/x-time":
			return null;
		case "application/x-boolean":
			return value ? "true" : "false";
		case "application/x-choice":
			if (cp.$value && cp.$value.$type === "application/x-string") {
				return "'" + value.replace(/'/g, "\\'") + "'";
			} else return value + "";
			break;
		case "application/x-reference":
			if (typeof value != "string") {
				value = filter.value;
			}
			cv = (value || '').replace(/'/g, "\\'");
			if (filter.operator === "like") cv = '%' + cv + '%';
			else if (filter.operator === "like_s") cv = cv + '%';
			cv = "'" + cv + "'";
			return cv;
		default:
			break;

	}
	return null;
};
