// syracuse-ui/lib/desktop/field/collection/builder/filter/quickFilterRecord.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Article = require("syracuse-ui/lib/desktop/article/article").Article;
var QuickFilterCell = require('./quickFilterCell').QuickFilterCell;

function QuickFilterRecord(){
}

exports.QuickFilterRecord = helpers.defineClass(QuickFilterRecord, Article, {
    loadRecord: function(list){
        this.$facet = "$filter";
        this.$isEditMode = true;
        this.$prototype = list.$prototype.$item;
        this.$prototype.$localization = list.getPage().$prototype.$localization;
        document.itemFactory.initializeItem(this, {
            $layout: {
                $items: list.$item.$layout.$items
            }
        }, list);
        this.loadBox();
    },
    _appendCell: function(slot){
        var td = document.createElement("td");
        td.className = "s-list-filter-cell";
        return slot.appendChild(td);
    },
    appendEmptyCell: function(slot){
        return this._appendCell(slot || this.$$item[0]);
    },
    appendFlagCell: function(slot){
        var td = this.appendEmptyCell(slot);
        td.className += " s-list-filter-flag";
        td.style.backgroundImage = "url('" + document.site.$item.$iconPath + "filter/s-filter-flag.png')";
        return td;
    },
    appendFieldCell: function($item, $field){
        var td = this._appendCell(this.boxParent.$item.$format == "vgrid" ? this.boxParent.builder.table.cols[$item.$bind].slot : this.$$item[0]);
        if ($field.$capability.filter) {
            (this._filters[$item.$bind] = new QuickFilterCell()).render(this, $(td), $field, $item.$bind);
        }
    },
    drawBox: function(){
        this._filters = {};
        if (this.boxParent.$item.$format != "vgrid") {
            if (this.$$item) {
                this.$$item.empty()[0];
            }
            else {
                var item = document.createElement("tr");
                this.boxParent.builder.table._thead.appendChild(item);
                this.$$item = $(this.boxParent.builder.table._thead.appendChild(item));
            }
        }
    },
    _buildFilter: function(){
        var self = this;
        var values = [];
        var params = {};
        Object.keys(self._filters).forEach(function($bind){
            var value = self._filters[$bind].getValue();
            if (value) {
                values.push(value);
            }
        });
        params.filter = values.join(" and ");
        this.boxParent.fetch(params);
    },
    onNotifyRecordChange: function(value, $bind){
        this.dataset[$bind] = value;
        var filter = this._filters[$bind];
        if (value == null || value == "") {
            if (filter && filter.operator !== "none") {
                if (filter.operator == "between") {
                    var field = this.record.boundFields[filter.$bind][0];
                    var fieldHigh = this.record.boundFields[filter.$bind + "$High"][0];
                    if (field.currentValue == fieldHigh.currentValue) {
                        filter.setOperator("none");
                    }
                }
                else {
                    filter.setOperator("none");
                }
            }
        }
        else {
            if (filter) {
                //for high this._filters[$bind] is not defined
                filter.ensureEqual();
            }
        }
        this._buildFilter();
        return null;
    },
    dispose: function(){
        var self = this;
        if (self._filters) {
            Object.keys(self._filters).forEach(function($bind){
                document.controller.disposeObject(self._filters[$bind]);
            });
            delete self._filters;
        }
        Article.prototype.dispose.call(this);
    },
    // FDB 
    fusionGetInputValues: function(){
        var self = this;
        var res = [];
        Object.keys(self._filters).forEach(function($bind){
            res.push(self._filters[$bind].fusionGetInputValue() || "");
        });
        return res;
    }
});
