// syracuse-ui/lib/tests/misc/printer/selectPrinterTest.js
"use strict";
var helpers = require("syracuse-core/lib/helpers");
var TestPage = require("syracuse-ui/lib/tests/testPage").TestPage;
var selectPrinterProto = require("syracuse-ui/lib/tests/misc/printer/selectPrinterProto");

function SelectPrinterTest() {}

exports.SelectPrinterTest = helpers.defineClass(SelectPrinterTest, TestPage, {
	_initiliazeTestPage: function() {
		document.site.deltaManager.applyObjectDelta(this, this.$prototype, selectPrinterProto.$prototype, true);
		this.$prototype.$facet = "$edit";
		this.$prototype.$actions.$openDialog = {
			$title: "openDialog"
		};
		this.$item.$layout = {
			$items: [{
					$category: "link",
					$bind: "$openDialog"
				},
				selectPrinterProto.$article.$layout
			]
		};
	},
	loadBox: function() {
		TestPage.prototype.loadBox.call(this);
		this.fillPage(this);
	},
	notifyActionChange: function(menuItem, target, value) {
		if (value.$openDialog) {
			var self = this;
			document.site.dialogManager.openModal(self, {
				$itemPage: {
					$facet: "edit",
					$representation: {
						$prototype: selectPrinterProto.$prototype,
						$article: selectPrinterProto.$article
					},
					onAfterLoadPage: function(page) {
						self.fillPage(page);
					}
				}
			});
		}
	},
	fillPage: function(page) {
		page.applyChange({
			$links: {
				"1186": {
					"$isHidden": false,
					"$isDisabled": true
				}
			}
		});
		page.applyChange({
			"printers": [{
				"$uuid": 0,
				"$serverIndex": 0,
				"$rawV": {
					"name": {},
					"port": {},
					"driver": {},
					"comment": {},
					"location": {},
					"$capabilities": {}
				},
				"name": "\\\\172.28.8.4\\KONICA Grenoble",
				"driver": "KONICA MINOLTA 423SeriesPCL",
				"port": "172.28.8.152",
				"location": "",
				"comment": "",
				"$capabilities": "collate,copies,orientation,duplex"
			}, {
				"$uuid": 1,
				"$serverIndex": 1,
				"$rawV": {
					"name": {},
					"port": {},
					"driver": {},
					"comment": {},
					"location": {},
					"$capabilities": {}
				},
				"name": "TEC B-SX4",
				"driver": "TEC B-SX4",
				"port": "FILE:",
				"location": "",
				"comment": "",
				"$capabilities": "copies,orientation"
			}, {
				"$uuid": 2,
				"$serverIndex": 2,
				"$rawV": {
					"name": {},
					"port": {},
					"driver": {},
					"comment": {},
					"location": {},
					"$capabilities": {}
				},
				"name": "Microsoft XPS Document Writer",
				"driver": "Microsoft XPS Document Writer",
				"port": "XPSPort:",
				"location": "",
				"comment": "",
				"$capabilities": "orientation"
			}, {
				"$uuid": 3,
				"$serverIndex": 3,
				"$isSelected": true,
				"$rawV": {
					"name": {},
					"port": {},
					"driver": {},
					"comment": {},
					"location": {},
					"$capabilities": {}
				},
				"name": "Fax",
				"driver": "Microsoft Shared Fax Driver",
				"port": "SHRFAX:",
				"location": "",
				"comment": "",
				"$capabilities": "orientation"
			}, {
				"$uuid": 4,
				"$serverIndex": 4,
				"$rawV": {
					"name": {},
					"port": {},
					"driver": {},
					"comment": {},
					"location": {},
					"$capabilities": {}
				},
				"name": "Envoyer a  OneNote 2010",
				"driver": "Send To Microsoft OneNote 2010 Driver",
				"port": "nul:",
				"location": "",
				"comment": "",
				"$capabilities": "orientation"
			}, {
				"$uuid": 5,
				"$serverIndex": 5,
				"$rawV": {
					"name": {},
					"port": {},
					"driver": {},
					"comment": {},
					"location": {},
					"$capabilities": {}
				},
				"name": "\\\\172.28.8.4\\HP3390-grenoble",
				"driver": "HP Universal Printing PCL 5",
				"port": "172.28.8.151",
				"location": "",
				"comment": "",
				"$capabilities": "collate,copies,orientation,duplex"
			}, {
				"$uuid": 6,
				"$serverIndex": 6,
				"$rawV": {
					"name": {},
					"port": {},
					"driver": {},
					"comment": {},
					"location": {},
					"$capabilities": {}
				},
				"name": "Test BID cmi1",
				"driver": "Canon Inkjet iP100 series",
				"port": "LPT1:",
				"location": "",
				"comment": "",
				"$capabilities": "collate,copies,orientation,duplex"
			}, {
				"$uuid": 7,
				"$serverIndex": 7,
				"$rawV": {
					"name": {},
					"port": {},
					"driver": {},
					"comment": {},
					"location": {},
					"$capabilities": {}
				},
				"name": "\\\\ECRDECHENAUX\\Datamax M-4206",
				"driver": "Datamax M-4206",
				"port": "LPT1:",
				"location": "",
				"comment": "",
				"$capabilities": "copies,orientation"
			}, {
				"$uuid": 8,
				"$serverIndex": 8,
				"$rawV": {
					"name": {},
					"port": {},
					"driver": {},
					"comment": {},
					"location": {},
					"$capabilities": {}
				},
				"name": "\\\\sageimp\\begonia",
				"driver": "HP LaserJet 4050 Series PCL 5e",
				"port": "IP_172.16.22.1",
				"location": "COLISEE",
				"comment": "172.16.22.1 Laserjet 4050",
				"$capabilities": "collate,copies,orientation"
			}, {
				"$uuid": 9,
				"$serverIndex": 9,
				"$rawV": {
					"name": {},
					"port": {},
					"driver": {},
					"comment": {},
					"location": {},
					"$capabilities": {}
				},
				"name": "\\\\SageImpX64\\Perce-Neige N&B",
				"driver": "HP Color LaserJet 4700 PCL 6 (Black)",
				"port": "perce-neige.sagefr.adinternal.com",
				"location": "COLISEE",
				"comment": "172.16.22.4 HP Laserjet 4700 couleur",
				"$capabilities": "collate,copies,orientation,duplex"
			}, {
				"$uuid": 10,
				"$serverIndex": 10,
				"$rawV": {
					"name": {},
					"port": {},
					"driver": {},
					"comment": {},
					"location": {},
					"$capabilities": {}
				},
				"name": "\\\\vpmon-it\\Pommier B4",
				"driver": "KONICA MINOLTA 1050 VXL",
				"port": "172.27.0.23 Pommier",
				"location": "FR/Montreuil/Sage France",
				"comment": "172.27.0.23 Konica BizHub Pro 1050 Bac 4",
				"$capabilities": "copies,orientation,duplex"
			}],
			"$isPartialDelta": false
		});
	},
	applyFieldMetaData: function($bind, value, $field) {

	}
});
