// msoffice/lib/excel/contract/entities/excelDatasource.js
"use strict"

exports.entity = {
	$title: "Datasources",
	$entityName: "excelDatasource",
	$pluralName: "excelDatasources",
	$capabilities: "edit,delete",
	$properties: {
		description: {
			$title: "Description"
		},
		dsName: {
			$title: "Name"
		},
		title: {
			$title: "Title"
		},
		endpoint: {
			$title: "Endpoint"
		},
		entity: {
			$title: "Entity"
		},
		representation: {
			$title: "Representation"
		},
		fetchAll: {
			$title: "Fetch all",
			$type: "boolean"
		},
		fetchLimit: {
			$title: "Fetch limit",
			$type: "integer",
			$default: 1000,
			$isDefined: function(instance) {
				return !instance.fetchAll;
			}
		},
		filter: {
			$title: "Filter",
			$type: "filter",
			$isDisabled: function(instance) {
                return instance.entity;
			},
			$applyMeta: function(instance) {
				if(instance.dataBaseUrl && instance.representation) {
					instance.$properties = instance.$properties || {};
					var f = instance.$properties.filter = instance.$properties.filter || {};
					f.$links = f.$links || {};
					f.$links.$prototype = f.$links.$prototype || {};
					f.$links.$prototype.$url = "{dataBaseUrl}/$prototypes('{representation}.$query')";
				}
			}
		}
	},
	$relations: {
		endpointRef: {
			$title: "Endpoint",
			$isMandatory: true,
			$value: "{description}",
			$baseUrl: document.site.excelDocument.syracuseBaseUrl,
			$type: "endPoints",
			$representation: "endPoint",
			$key: "{dataset}", // key MUST be different of "{$uuid}" to have the lookup fetch all the record
			$propagate: function(resource, val) {
				if(val) {
					resource.applicationId = val.applicationRef.$uuid;
					resource.endpoint = val.dataset;
					resource.dataset = val.dataset;
					resource.dataBaseUrl = ["/sdata", val.application, val.contract, val.dataset].join("/");
				} else {
					resource.applicationId = "";
					resource.endpoint = "";
					resource.dataset = "";
					resource.databaseUrl = "";
				}
			}
		},
		menuItem: {
			$title: "Datasource",
			$value: "{description}",
			$baseUrl: document.site.excelDocument.syracuseBaseUrl,
			$type: "menuItems",
			$representation: "menuItem",
			$key: "{description}", // key MUST be different of "{$uuid}" to have the lookup fetch all the record
			$isDisabled: function(instance) {
				return instance.endpoint == null;
			},
			$lookup: {
				$filter: "(application eq \"{applicationId}\") and (linkType eq \"$representation\") and (facet eq \"$query\")"
			},
			$propagate: function(resource, val) {
				if(val) {
					resource.entity = val.representationRef.entity;
					resource.representation = val.representationRef.representation;
				} else {
					resource.entity = "";
					resource.representation = "";
				}
			}
		},
		representationRef: {
			$title: "Representation",
			$type: "representationProxy",
			$key: "{respresentation}", // key MUST be different of "{$uuid}" to have the lookup fetch all the record
			isChild: true,
			$lookup: {
				parameters: "dataset={endpoint}"
			},
			$propagate: function(resource, val) {
				if(val) {
					resource.representation = val.representation;
					resource.entity = val.entity;
					resource.title = val.title;
				} else {
					resource.representation = "";
					resource.entity = "";
				}
			},
			$isDisabled: function(instance) {
				return instance.endpoint == null;
			}
		},
		orderBys: {
			$title: "Order bys",
			$isArray: true,
			$isPlural: true,
			$baseUrl: document.site.excelDocument.syracuseBaseUrl,
			$value: "{name}",
			$itemTitle: "{title}",
			$properties: { title: { $type: "application/x-string" }, name: { $type: "application/x-string" }},
			$type: "excelOrderAttr",
			$capabilities: "insert,append,delete,reorder",
			$select: {
				$title: "Attributes",
				$url: document.site.excelDocument.syracuseBaseUrl + "/lookupEntityAttrs?representation=lookupEntityAttr.$select",
				$parameters: "dataset={endpoint}&entity={entity}"
			}
		}
	},
	$actions: {
		$save: {
			$title: "Apply",
			$links: {
				$details: {
					$title: "Ok",
					$url: "/"
				}
			}
		}
	},
	$links: {
		refresh: {
			$title: "Refresh",
			$url: "excel://excelDatasources('{$uuid}')/$service/refresh?representation=excelDatasource.$details",
			$method: "POST",
			$facets: "$query,$details"
		}
	},
	$queryLinks: {
		browseMenu: {
			$title: "Predefined datasources",
			$type: "application/json; vnd-sage=syracuse",
			$url: "excel://excelDatasources/$template/$workingCopies?representation=excelBrowseDsMenuItem.$edit",
			$method: "POST"
		},
		browseEntity: {
			$title: "Data model browse",
			$type: "application/json; vnd-sage=syracuse",
			$url: "/sdata/syracuse/collaboration/syracuse/excelBrowseDatasources/$template/$workingCopies?representation=excelBrowseDsEntity.$edit",
			$method: "POST"
		},
		refreshAll: {
			$title: "Refresh all",
			$url: "excel://excelDatasources/$service/refreshAll?representation=excelDatasource.$query",
			$method: "POST"
		}
	},
	// standard functions, MUST implement
	$fetch: function(id) {
		return document.site.excelDocument.getDatasourcesResource(id);
	}
}
