// msoffice/lib/excel/contract/entities/excelSIParams.js
"use strict";

exports.entity = {
	$title: "Sage Intelligence",
	$entityName: "excelSIParam",
	$pluralName: "excelSIParams",
	$properties: {
		ledType: {
			$title: "Ledger type (det)",
			$isHidden: true,
			$type: "integer",
			$enum: [{
				$value: 0,
				$title: "undefined"
			}, {
				$value: 1,
				$title: "Social"
			}, {
				$value: 2,
				$title: "Analytic"
			}]
		}
	},
	$relations: {
		endpoint: {
			$title: "Endpoint",
			$isMandatory: true,
			$value: "{description}",
			$baseUrl: syra_site.excelDocument.syracuseBaseUrl,
			$type: "endPoints",
			$representation: "endPoint",
			$lookup: {
				$filter: "(protocol eq \"x3\")"
			}
		},
		coa: {
			$title: "Chart of accounts",
			$isMandatory: true,
			$value: "{COA}",
			$type: "GCOA",
			$representation: "GCOA",
		},
		let: {
			$title: "Ledger type",
			$isMandatory: true,
			$value: "{GCM}",
			$type: "GACM",
			$representation: "GACM"
		},
		/*		pyramidSoc: {
			$title: "Pyramid",
			$isMandatory: false,
			$value: "{GYM}",
			$type: "GYM",
			$representation: "GYM"
		},*/
		pyramidAna: {
			$title: "Pyramid",
			$isMandatory: false,
			$value: "{PYM}",
			$type: "GCCEPYM",
			$representation: "GCCEPYM"
		},
		groupSoc: {
			$title: "Group",
			$value: "{GRY}",
			$type: "GRY",
			$representation: "GRY"
		},
		groupAna: {
			$title: "Group",
			$value: "{CRY}",
			$type: "CRY",
			$representation: "CRY"
		}
	}
};
