// syracuse-collaboration/lib/representations/dashboardDef.js
"use strict"

exports.representation = {
	$entityName: "dashboardDef",
	$facets: {
		$details: {
			$layout: {
		        "$layoutType": "stack",
		        "$items": [
		            {
		                "$category": "section",
		                "$title": "Informations",
		                "$layout": {
		                    "$items": [
		                        {
		                            "$bind": "title"
		                        },
		                        {
		                            "$bind": "description"
		                        },
		                        {
		                            "$bind": "dashboardName"
		                        }
		                    ],
		                    "$layoutType": "stack"
		                },
		                "$opened": true
		            },
		            {
		                "$category": "section",
		                "$title": "Variants",
		                "$layout": {
		                    "$items": [
		                        {
		                            "$bind": "variants",
		                            "$isQuickFilter": true,
                                    "$isTitleHidden": true,
		                            "$layout": {
		                                "$items": [
		                                    {
		                                        "$bind": "title"
		                                    },
		                                    {
		                                        "$bind": "description"
		                                    },
		                                    {
		                                        "$bind": "allApplications"
		                                    },
		                                    {
		                                        "$bind": "application"
		                                    },
		                                    {
		                                        "$bind": "$creUser"
		                                    },
		                                    {
		                                        "$bind": "$updUser"
		                                    },
		                                    {
		                                        "$bind": "$creDate"
		                                    },
		                                    {
		                                        "$bind": "$updDate"
		                                    }
		                                ]
		                            },
		                            "$cards": {
		                                "$layout": {
		                                    "$layoutType": "tabs",
		                                    "$items": [
		                                        {
		                                            "$category": "section",
		                                            "$title": "Vignettes",
		                                            "$layout": {
		                                                "$items": [
		                                                    {
		                                                        "$bind": "vignettes",
		                                                        "$isTitleHidden": true,
		                    				                    "$format": "cards",
		                    				                    "$cardsByRowCount": 3,
		                                                        "$layout": {
		                                                        	"$items": [{
		                                                        		"$bind": "portlet"
		                                                        	}, {
		                                                        		"$bind": "allEndpoints"
		                                                        	}, {
		                                                        		"$bind": "endpoint"
		                                                        	}]
		                                                        }
		                                                    }
		                                                ]
		                                            }
		                                        },
		                                        {
		                                            "$category": "section",
		                                            "$title": "Applies to roles",
		                                            "$layout": {
		                                                "$items": [
		                                                    {
		                                                        "$bind": "roles",
		                                                        "$isTitleHidden": true,
		                                                        "$isQuickDesignerEnabled": false,
		                                                        "$isPagerHidden": true,
		                                                        "$selectMode": "multi",
		                                                        "$layout": {
		                                                            "$items": [
		                                                                {
		                                                                    "$bind": "description"
		                                                                }
		                                                            ]
		                                                        }
		                                                    }
		                                                ],
		                                                "$layoutType": "stack"
		                                            },
		                                            "$opened": true
		                                        },
		                                        {
		                                            "$category": "section",
		                                            "$title": "Applies to users",
		                                            "$layout": {
		                                                "$items": [
		                                                    {
		                                                        "$bind": "users",
		                                                        "$isTitleHidden": true
		                                                    }
		                                                ]
		                                            }
		                                        },
		                                        {
		                                            "$category": "section",
		                                            "$title": "Applies to endpoints",
		                                            "$layout": {
		                                                "$items": [
		                                                    {
		                                                        "$bind": "endpoints",
		                                                        "$isTitleHidden": true
		                                                    }
		                                                ]
		                                            }
		                                        }
		                                    ]
		                                },
		                                "$category": "section"
		                            }
		                        }
		                    ],
		                    "$layoutType": "stack"
		                }
		            }
		        ]
			}
		},
		$edit: {
			$copy: "$details"
		}
	}
}
