// syracuse-collaboration/lib/representations/documentExcel.js
exports.representation = {
	$entityName: "document",
	$facets: {
		$details: {
			$options: {
//				$notFoundUrl: "{$baseUrl}/excelDocumentPublish?representation=excelDocumentPublish.$query"
			},
			$prototype: {
				$links: {
					$query: {
						$isHidden: true
					},
					$delete: {
						$isHidden: true
					}
				},
				$actions: {
					$saveDocument: {
						$title: "Save document to X3"
					}
				}
			},
			$layout: {
                $items: [{
                    $layoutType: "columns",
                    $layoutSubType: "50,50",
                    $items: [{
						$items: [{
							$bind: "description"
						}, {
							$bind: "documentDate"
						}]
					}, {
						$items: [{
							$bind: "owner"
						}, {
							$bind: "isReadOnly"
						}]
					}]
				}, {
                    $layoutType: "stack",
					$items: [{
						$bind: "teams"
					}, {
						$bind: "tags"
					}]
				}]
			}
		},
		$edit: {
			$prototype: {
				$links: {
					$query: {
						$isHidden: true
					}
				},
				$actions: {
					$save: {
						$links: {
							$query: {
								$isHidden: true
							},
							$create: {
								$isHidden: true
							}
						}
					}
				}
			},
			$layout: {
				$copy: "$details"
			}
		},
		$query: {
			$prototype: {
				$links: {
					$create: {
						$title: "Publish"
					},
					$print: {
						$isHidden: true
					},
					$excel: {
						$isHidden: true
					}
				}
			},
            $layout: {
                $items: [{
                    $category: "section",
                    $layout: {
                        $items: [{
                            $bind: "$resources",
//                            $format: "excelDocumentCardList",
                            $format: "cardList",
                            $cardview: {
                            	$layout: {
                            		$items: [{
	                                   	"$category":"block",
	                                    "$expression":"{description} published by {owner} on {documentDate}, shared with {teams}"
                            		}]
                            	}
                            }
                        }]
                    }
                }]
            }
		}
	}
}
