// syracuse-collaboration/lib/representations/endpoint.js
exports.representation = {
    $entityName: "endPoint",
    $facets: {
        $details: {
            $layout: {
                $items: [{
                    $title: "{@InformationsSectionTitle}",
                    $category: "block",
                    $layout: {
                        $items: [{
                            $bind: "dataset"
                        },{
                            $bind: "description"
                        }]
                    }
                }, {
                    $title: "{@LocationSectionTitle}",
                    $category: "block",
                    $layout: {
                        $items: [{
                            $bind: "applicationRef"
                        }]
                    }
                }, {
                    $title: "{@ServerSectionTitle}",
                    $category: "block",
                    $layout: {
                        $items: [{
                            $bind: "x3server"
                        }, {
                            $bind: "x3ServerFolder"
                        }, {
                            $bind: "databaseDriver"
                        }, {
                            $bind: "databaseHost"
                        }, {
                            $bind: "databasePort"
                        }, {
                        	$bind: "transitionWebServer"
                        }]
                    }
                }, {
                    $title: "{@AdministrationSectionTitle}",
                    $category: "block",
                    $layout: {
                        $items: [{
                            $bind: "groups"
                        }]
                    }
                }, {
                    $title: "{@OptionsSectionTitle}",
                    $category: "block",
                    $layout: {
                        $items: [{
                            $bind: "enableSearch"
                        }]
                    }
                }]
            }
        },
        $edit: {
            $copy: "$details"
        }
    }
}
