// syracuse-collaboration/lib/representations/exportProfile.js
exports.representation = {
    $entityName: "exportProfile",
    $facets: {
        $details: {
            $layout: {
                $items: [{
                    $title: "{@InformationsSectionTitle}",
                    $category: "block",
                    $layout: {
                        $items: [{
                            $bind: "description"
                        }]
                    }
                }, {
                    $title: "{@AdministrationSectionTitle}",
                    $category: "block",
                    $layout: {
                        $items: [{
                            $bind: "application"
                        }, {
                            $bind: "endpoint"
                        }]
                    }
                }]
            }
        },
        $edit: {
            $copy: "$details"
        },
        $query: {
            $layout: {
                $items: [{
                    $category: "section",
                    $layout: {
                        $items: [{
                            $bind: "$resources"
                        }]
                    }
                }]
            }
        }
    }
}
