// syracuse-collaboration/lib/representations/group.js
"use strict";
exports.representation = {
    $entityName: "group",
    $facets: {
        $details: {
            $layout: {
                $layoutType: "row",
                $items: [{
                    $items: [{
                        $title: "{@InformationsSectionTitle}",
                        "$category": "section",
                        $layout: {
                            $items: [{
                                $bind: "description"
                            }, {
                                $bind: "parent"
                            }, {
                                $bind: "children"
                            }]
                        }
                    }, {
                        $title: "{@AdministrationSectionTitle}",
                        "$category": "section",
                        $layout: {
                            $items: [{
                                $bind: "role"
                            }, {
                                $bind: "users"
                            }, {
                                $bind: "endPoints"
                            }]
                        }
                    }]
                }, {
                    $items: [{
                        $title: "{@ExplorerSectionTitle}",
                        "$category": "section",
                        $layout: {
                            $items: [{
                                $bind: "explorer"
                            }]
                        }
                    }]
                }]
            }
        },
        $edit: {
            $copy: "$details"
        },
        $query: {
            $layout: {
                $items: [{
                    $category: "section",
                    $layout: {
                        $items: [{
                            $bind: "$resources",
                            $layout: {
                                $items: [{
                                    $bind: "description"
                                }, {
                                    $bind: "role"
                                }, {
                                    $bind: "parent"
                                }]
                            }
                        }]
                    }
                }]
            }
        }
    }
}
