// syracuse-collaboration/lib/representations/groupMob.js
exports.representation = {
    $entityName: "group",
    $facets: {
        $details: {
            $layout: {
                $items: [{
                    $title: "Informations",
                    $category: "block",
                    $layout: {
                        $items: [{
                            $bind: "description"
                        }, {
                            $bind: "parent"
                        }, {
                            $bind: "children"
                        }]
                    }
                }, {
                    $title: "Administration",
                    $category: "block",
                    $layout: {
                        $items: [{
                            $bind: "role"
                        }, {
                            $bind: "users"
                        }, {
                            $bind: "endPoints"
                        }]
                    }
                }]
            }
        },
        $edit: {
            $copy: "$details"
        },
        $query: {
            $layout: {
                $items: [{
                    $category: "section",
                    $layout: {
                        $items: [{
                            $bind: "$resources",
                            $format: "cardList",
                            $cardView: {
                                $layout: {
                                    $layoutType: "stack",
                                    $items: [{
                                        $bind: "description",
                                        $isTitleHidden: true
                                    }, {
                                        $bind: "role",
                                        $isTitleHidden: true
                                    }]
                                }
                            }
                        }]
                    }
                }]
            }
        }
    }
}
