// syracuse-collaboration/lib/representations/host.js
"use strict";
exports.representation = {
    $entityName: "host",
    $facets: {
        $details: {
            $layout: {
                $items: [{
                    "$category": "section",
                    $layout: {
                        $items: [{
                            $bind: "hostname"
                        }, {
                            $bind: "nannyPort"
                        }, {
                            $bind: "children"
                        }, {
                            $bind: "active"
                        }, {
                            $bind: "started"
                        }, {
                            $bind: "status"
                        }, {
                            $bind: "version"
                        }, {
                            $bind: "domain"
                        }, {
                            $bind: "pid"
                        }, {
                            $bind: "respawnCount"
                        }, {
                            $bind: "respawnTime"
                        }, {
                            $bind: "returnRequestTimeout"
                        }, {
                            $bind: "childInformation",
                            $rows: 20
                        }]
                    }
                }]
            }
        },
        $edit: {
            $copy: "$details"
        },
        $query: {
            $layout: {
                $items: [{
                    $category: "section",
                    $layout: {
                        $items: [{
                            $bind: "$resources",
                            $layout: {
                                $items: [{
                                    $bind: "hostname"
                                }, {
                                    $bind: "nannyPort"
                                }, {
                                    $bind: "children"
                                }, {
                                    $bind: "active"
                                }, {
                                    $bind: "started"
                                }, {
                                    $bind: "status"
                                }, {
                                    $bind: "version"
                                }, {
                                    $bind: "domain"
                                }]
                            }
                        }]
                    }
                }]
            }
        }
    }
}
