// syracuse-collaboration/lib/representations/ldap.js
"use strict";
exports.representation = {
	resources: function() {
		var locale = require("syracuse-core/lib/locale");
		var globals = require('streamline/lib/globals');
		return locale.resources(module, globals.context && globals.context.sessionLocale)();
	},
	$entityName: "ldap",
	$facets: {
		$details: {
			$layout: {
				$layoutType: "row",
				$items: [{
					$items: [{
						$title: "{@ServerSectionTitle}",
						"$category": "section",
						$layout: {
							$items: [{
								$bind: "name"
							}, {
								$bind: "displayName"
							}, {
								$bind: "active"
							}, {
								$bind: "url"
							}, {
								$bind: "adminDn"
							}, {
								$bind: "adminPassword"
							}, {
								$bind: "searchBase"
							}, {
								$bind: "cacerts"
							}]
						}
					}, {
						$title: "{@UsersSectionTitle}",
						"$category": "section",
						$layout: {
							$items: [{
								$bind: "searchFilter"
							}, {
								$bind: "syncSerachFilter"
							}, {
								$bind: "authenticationNameMapping"
							}, {
								$bind: "firstNameMapping"
							}, {
								$bind: "lastNameMapping"
							}, {
								$bind: "emailMapping"
							}, {
								$bind: "photoMapping"
							}, {
								$bind: "userGroupNameMapping"
							}, {
								$bind: "users"
							}]
						}
					}]
				}, {
					$items: [{
						$title: "{@GroupsSectionTitle}",
						"$category": "section",
						$layout: {
							$items: [{
								$bind: "groupSearchFilter",
							}, {
								$bind: "groupNameMapping"
							}]
						}
					}]
				}]
			}
		},
		$edit: {
			$copy: "$details"
		}
	
	}
};
