// syracuse-collaboration/lib/representations/license.js
exports.representation = {
    $entityName: "license",
    $facets: {
        $details: {
            $layout: {
                $items: [{
                    "$category": "section",
                    $layout: {
                        $items: [{
                            $bind: "path"
                        }, {
                            $bind: "content",
                            $rows: 20
                        }]
                    }
                }]
            }
        },
        $edit: {
            $copy: "$details"
        },
        $query: {
            $layout: {
                $items: [{
                    $category: "section",
                    $layout: {
                        $items: [{
                            $bind: "$resources",
                            $layout: {
                                $items: [{
                                    $bind: "content"
                                }]
                            }
                        }]
                    }
                }]
            }
        }
    }
}
