// syracuse-collaboration/lib/representations/menuItem.js
exports.representation = {
	entityName: "menuItem",
	$facets: {
		$details: {
			$layout: {
				$items: [{
					$category: "section",
					$layout: {
						$items: [{
							$layoutType: "columns",
							$layoutSubType: "50,50",
							$items: [{
								$items: [{
									$category: "block",
									$title: "Informations",
									$layout: {
										$items: [{
											$bind: "code"
										}, {
											$bind: "title"
										}, {
											$bind: "description"
										}, {
											$bind: "linkType"
										}]
									}
								}, {
									$category: "block",
									$title: "Content",
									$layout: {
										$items: [{
											$bind: "application"
										}, {
											$bind: "endpoint"
										}, {
											$bind: "entity"
										}, {
											$bind: "representation"
										}, {
											$bind: "facet"
										}, {
											$bind: "dashboard"
										}, {
											$bind: "fusionFunction"
										}, {
											$bind: "fusionKey"
										}, {
											$bind: "externalUrl"
										}]
									}
								}]
							}, {
								$items: [{
									$category: "block",
									$title: "Parameters",
									$layout: {
										$items: [{
											$bind: "keyParameter"
										}, {
											$bind: "parameters"
										}]
									}
								}, {
									$category: "block",
									$title: "Display",
									$layout: {
										$items: [{
											$bind: "target"
										}, {
											$bind: "menus"
										}]
									}
								}, {
									$category: "block",
									$title: "Classification",
									$layout: {
										$items: [{
											$bind: "categories"
										}]
									}
								}]
							}]
						}]
					}
				}]
			}
		},
		$edit: {
			$copy: "$details"
		},
		$query: {
			$layout: {
				$items: [{
					$category: "section",
					$layout: {
						$items: [{
							$bind: "$resources",
							$layout: {
								$items: [{
									$bind: "title",
								}, {
									$bind: "description"
								}, {
									$bind: "linkType"
								}, {
									$bind: "application"
								}, {
									$bind: "endpoint"
								}]
							}
						}]
					}
				}]
			}
		},
		$lookup: {
			$copy: "$query"
		},
		$select: {
			$copy: "$query"
		}
	}
}
