// syracuse-collaboration/lib/representations/menuItem.js
exports.representation = {
    $entityName: "menuItem",
    $facets: {
        $details: {
            $layout: {
                $layoutType: "columns",
                $layoutSubType: "50,50",
                $items: [{
                    $items: [{
                        $category: "section",
                        $title: "{@InformationsSectionTitle}",
                        $layout: {
                            $items: [{
                                $bind: "code"
                            }, {
                                $bind: "title"
                            }, {
                                $bind: "description"
                            }, {
                                $bind: "linkType"
                            }, {
                                $bind: "icon"
                            }, {
                                $bind: "applicationMenu"
                            }]
                        }
                    }, {
                        $category: "section",
                        $title: "{@ContentSectionTitle}",
                        $layout: {
                            $items: [{
                                $bind: "application"
                            }, {
                                $bind: "endpoint"
                            }, {
                                $bind: "entity"
                            }, {
                                $bind: "representation"
                            }, {
                                $bind: "facet"
                            }, {
                                $bind: "dashboard"
                            }, {
                                $bind: "fusionFunction"
                            }, {
                                $bind: "fusionKey"
                            }, {
                                $bind: "externalUrl"
                            }]
                        }
                    }]
                }, {
                    $items: [{
                        $category: "section",
                        $title: "{@ParametersSectionTitle}",
                        $layout: {
                            $items: [{
                                $bind: "keyParameter"
                            }, {
                                $bind: "parameters"
                            }]
                        }
                    }, {
                        $category: "section",
                        $title: "{@DisplaySectionTitle}",
                        $layout: {
                            $items: [{
                                $bind: "target"
                            }, {
                                $bind: "menus"
                            }]
                        }
                    }, {
                        $category: "section",
                        $title: "{@ClassificationSectionTitle}",
                        $layout: {
                            $items: [{
                                $bind: "categories"
                            }]
                        }
                    }]
                }]
            
            }
        },
        $edit: {
            $copy: "$details"
        },
        $query: {
            $layout: {
                $items: [{
                    $category: "section",
                    $layout: {
                        $items: [{
                            $bind: "$resources",
                            $layout: {
                                $items: [{
                                    $bind: "title",
                                }, {
                                    $bind: "description"
                                }, {
                                    $bind: "linkType"
                                }, {
                                    $bind: "representation"
                                }, {
                                    $bind: "fusionFunction"
                                }, {
                                    $bind: "dashboard"
                                }, {
                                    $bind: "application"
                                }, {
                                    $bind: "endpoint"
                                }]
                            }
                        }]
                    }
                }]
            }
        },
        $lookup: {
            $copy: "$query"
        },
        $select: {
            $copy: "$query"
        }
    }
}
