// syracuse-collaboration/lib/representations/msoMailMergeDSSel.js
"use strict"

exports.representation = {
	$entityName: "msoMailMergeDSSel",
	$facets: {
		$query: {
			$prototype: {
				$links: {
					$query: {
						$isHidden: true
					},
					$delete: {
						$isHidden: true
					},
					$edit: {
						$isHidden: true
					},
					$create: {
						$isHidden: true
					},
					$print: {
						$isHidden: true
					},
					$excel: {
						$isHidden: true
					},
					$word: {
						$isHidden: true
					},
					
					selectDS: {
						$title: "Select datasource",
						$type: "application/json; vnd-sage=syracuse",
						$url: "/sdata/syracuse/collaboration/syracuse/msoMailMergeDSSels/$template/$workingCopies?representation=msoMailMergeDSSel.$edit",
						$method: "POST"
					},
					browseMenu: {
						$title: "Select by entity",
						$type: "application/json; vnd-sage=syracuse",
						$url: "/sdata/syracuse/collaboration/syracuse/msoMailMergeDSSels/$template/$workingCopies?representation=msoMailMergeDSSelEntity.$edit",
						$method: "POST"
					},
					browseEntity: {
						$title: "Select by menu",
						$type: "application/json; vnd-sage=syracuse",
						$url: "/sdata/syracuse/collaboration/syracuse/msoMailMergeDSSels/$template/$workingCopies?representation=msoMailMergeDSSelMenu.$edit",
						$method: "POST"
					}
				}
			},
			$layout: {
				$items: [
				]
			}
		},
		$edit: {
			
			$prototype: {
				$links: {
					$query: {
						$isHidden: true
					},
					$delete: {
						$isHidden: true
					},
					$edit: {
						$isHidden: true
					},
					$create: {
						$isHidden: true
					},
					$print: {
						$isHidden: true
					},
					$excel: {
						$isHidden: true
					},
					$word: {
						$isHidden: true
					},
				},
				$actions: {
					$save: {
						$title: "Attach datasource"
					}
				}

			},
			$layout: {
				$items: [{
					$bind: "datasource"
				}, {
					$bind: "title",
					$isEditMode: false					
				}, {
					$bind: "endpoint",
					$isEditMode: false
				}, {
					$bind: "menuItem",
					$isEditMode: false
				}, {
					$bind: "entity",
					$isEditMode: false					
				}, {
					$bind: "representation",
					$isEditMode: false					
				}, {
					$bind: "fetchCount",
					$isEditMode: false					
				}				]
			}
		}
	}
}
