// syracuse-collaboration/lib/representations/msoMailMergeDS.js
"use strict"

exports.representation = {
	$entityName: "msoMailMergeDS",
	$facets: {
		$edit: {
			$layout: {
				$items: [
				{
					$bind: "title"
				}, {
					$bind: "entity"
				}, {
					$bind: "representation"
				}, {
					$bind: "menuItem"
				}, {
					$bind: "fetchCount"
				}, {
					$bind: "endpoint"
				}
				]
			}
		},
        $query: {
            $layout: {
                $items: [{
                    $category: "section",
                    $layout: {
                        $items: [{
                            $bind: "$resources",
                            $layout: {
                                $items: [{
                                    $bind: "title"
                                }, {
                                	$bind: "entity"
                                }, {
                                	$bind: "representation"
                                }, {
                                	$bind: "menuItem"
                                }]
                            }
                        }]
                    }
                }]
            }
        },
	}
}
