// syracuse-collaboration/lib/representations/msoWordDocument.js
exports.representation = {
	$entityName: "document",
	$facets: {
		$query: {
			$prototype: {
				$links: {
					$query: {
						$isHidden: true
					},
					$create: {
						$isHidden: true
					},
					$delete: {
						$isHidden: true
					},
					$excel: {
						$isHidden: true
					},

					$print: {
						$isHidden: true
					},
					$wordmailmerge: {
						$isHidden: true
					},
					$wordreport: {
						$isHidden: true
					},
                    saveDocument: {
                        $title: "Save document to X3",
                        $type: "application/json;vnd.sage=syracuse",
                        $url: "/sdata/syracuse/collaboration/syracuse/documents/$template/$workingCopies?representation=msoWordDocument.$edit",
                        $method: "POST"
                    }
				},
			},
			$layout: {
                $items: [ ]
			}
		},
		$details: {
			$prototype: {
				$links: {
					$query: {
						$isHidden: true
					},
					$delete: {
						$isHidden: true
					}
				},
				$actions: {
					$saveDocument: {
						$title: "{@SaveActionTitle}"
					}
				}
			},
			$layout: {
                $items: [{
                    $layoutType: "columns",
                    $layoutSubType: "50,50",
                    $items: [{
						$items: [{
							$bind: "description"
						}, {
							$bind: "volume"
						}, {
							$bind: "documentDate"
						}]
					}, {
						$items: [{
							$bind: "owner"
						}, {
							$bind: "isReadOnly"
						}]
					}]
				}, {
                    $layoutType: "stack",
					$items: [{
						$bind: "teams"
					}, {
						$bind: "tags"
					}]
				}]
			}
		},
		$edit: {
			$prototype: {
				$links: {
					$query: {
						$isHidden: true
					},
					$print: {
						$isHidden: true
					},
					$wordreport: {
						$isHidden: true
					},
					$save: {
						$title: "Save to X3"
					}
				},
				$actions: {
					$save: {
						$links: {
							$query: {
								$isHidden: true
							},
							$create: {
								$isHidden: true
							}
						}
					}
				}
			},
			$layout: {
				$copy: "$details"
			}
		},
	}
}
