// syracuse-collaboration/lib/representations/pageData.js
exports.representation = {
    $entityName: "pageData",
    $facets: {
        $query: {
            $layout: {
                $items: [{
                    $category: "section",
                    $layout: {
                        $items: [{
                            $bind: "$resources",
                            $layout: {
                                $items: [{
                                    $bind: "title"
                                }, {
                                    $bind: "description"
                                }, {
                                    $bind: "application"
                                }, {
                                    $bind: "representation"
                                }, {
                                    $bind: "facet"
                                }, {
                                    $bind: "configuration"
                                }]
                            }
                        }]
                    }
                }]
            }
        },
        $details: {
            $layout: {
            	$items: [{
            		$category: "section",
            		$layout: {
                        $items: [{
                            "$category": "section",
                            $title: "{@InformationsSectionTitle}",
                            $layout: {
                                $items: [{
                                    $bind: "title"
                                }, {
                                    $bind: "description"
                                }, {
                                    $bind: "facet"
                                }, {
                                    $bind: "allApplications"
                                }, {
                                    $bind: "application"
                                }, {
                                    $bind: "representation"
                                }, {
                                	$bind: "dashboardName"
                                }, {
                                    $bind: "variant"
                                }]
                            }
                        }, {
                            "$category": "section",
                            $title: "{@ContentSectionTitle}",
                            $layout: {
                            	$items: [{
                                    $bind: "portlets",
                                    $layout: {
	                                    $items: [{
	                                    	$bind: "portlet"
	                                    }, {
	                                    	$bind: "allEndpoints"
	                                    }, {
	                                    	$bind: "endpoint"
	                                    }, {
	                                    	$bind: "titleTemplate"
	                                    }]
                                    }
                                }]
                            }
                        }, {
                            "$category": "section",
                            $title: "{@BindingSectionTitle}",
                            $layout: {
                                $items: [{
                                    $bind: "roles"
                                }, {
                                    $bind: "users"
                                }, {
                                	$bind: "endpoints"
                                }]
                            }
                        }, {
                            "$category": "section",
                            $title: "{@VersionSectionTitle}",
                            $layout: {
                                $items: [{
                                    $bind: "configuration"
                                }]
                            }
                        }]
            		}
            	}]
            }
        },
        $edit: {
            $copy: "$details"
        }
    }
}
