// syracuse-collaboration/lib/representations/portlet.js
"use strict"

exports.representation = {
	$entityName: "portlet",
	$facets: {
		$details: {
			$layout: {
				$items: [{
					$category: "section",
					$layout: {
						$items: [{
							$category: "block",
							$title: "Informations",
							$layout: {
								$items: [{
									$bind: "title"
								}, {
									$bind: "description"
								}, {
									$bind: "type"
								}]
							}
						}, {
							$category: "block",
							$title: "Filters",
							$layout: {
								$items: [{
									$bind: "application"
								}, {
									$bind: "endpoint"
								}]
							}
						}, {
							$category: "block",
							$title: "Content",
							$layout: {
								$items: [{
									$bind: "items"
								}, {
									$bind: "pageItem"
								}]
							}
						}, {
							$category: "block",
							$title: "Classification",
							$layout: {
								$items: [{
									$bind: "categories"
								}]
							}
						}]
					}
				}]
			}
		},
		$edit: {
			$copy: "$details"
		}
	}
}
