// syracuse-collaboration/lib/representations/portlet.js
"use strict"

exports.representation = {
    $entityName: "portlet",
    $facets: {
        $details: {
            $layout: {
                $items: [{
                    $category: "section",
                    $title: "Informations",
                    $layout: {
                        $items: [{
                            $bind: "title"
                        }, {
                            $bind: "description"
                        }, {
                            $bind: "type"
                        }]
                    }
                }, {
                    $category: "section",
                    $title: "Filters",
                    $layout: {
                        $items: [{
                            $bind: "application"
                        }, {
                            $bind: "endpoint"
                        }]
                    }
                }, {
                    $category: "section",
                    $title: "Content",
                    $layout: {
                        $items: [{
                            $bind: "items"
                        }, {
                            $bind: "pageItem"
                        }]
                    }
                }, {
                    $category: "section",
                    $title: "Classification",
                    $layout: {
                        $items: [{
                            $bind: "categories"
                        }]
                    }
                }]
            }
        
        },
        $edit: {
            $copy: "$details"
        }
    }
}
