// syracuse-collaboration/lib/representations/role.js
"use strict";
exports.representation = {
	resources: function() {
		var locale = require("syracuse-core/lib/locale");
		var globals = require('streamline/lib/globals');
		return locale.resources(module, globals.context && globals.context.sessionLocale)();
	},
	$entityName: "role",
	$facets: {
		$details: {
			$layout: {
				$layoutType: "row",
				$items: [{
					$items: [{
						$title: "{@InformationsSectionTitle}",
						$category: "section",
						$layout: {
							$items: [{
								$bind: "code"
							}, {
								$bind: "description"
							}]
						}
					}, {
						$title: "{@AdministrationSectionTitle}",
						$category: "section",

						$layout: {
							$items: [{
								$bind: "groups"
							}, {
								$bind: "badges"
							}, {
								$bind: "securityProfile"
							}, {
								$bind: "navigationPage"
							}, {
								$bind: "landingPages"
							}, {
								$bind: "boProfiles"
							}, {
								$bind: "mobileApplications"
							}]
						}
					}]
				}, {
					$items: [{
						$title: "{@ExplorerSectionTitle}",
						$category: "section",
						$layout: {
							$items: [{
								$bind: "explorer",
								$isTitleHidden: true
							}]
						}
					}]
				}]
			}
		},
		$edit: {
			$copy: "$details"
		}
	}
};
