// syracuse-collaboration/lib/representations/role.js
"use strict";
exports.representation = {
    $entityName: "role",
    $facets: {
        $details: {
            $layout: {
                $layoutType: "row",
                $items: [{
                    $items: [{
                        $title: "{@InformationsSectionTitle}",
                        $category: "section",
                        $layout: {
                            $items: [{
                                $bind: "description"
                            }, {
                                $bind: "functionProfile"
                            }]
                        }
                    }, {
                        $title: "{@AdministrationSectionTitle}",
                        $category: "section",
                        
                        $layout: {
                            $items: [{
                                $bind: "groups"
                            }, { 
                            	$bind: "badges"
                            }]
                        }
                    }]
                }, {
                    $items: [{
                        $title: "{@ExplorerSectionTitle}",
                        $category: "section",
                        $layout: {
                            $items: [{
                                $bind: "explorer",
                                $isTitleHidden: true
                            }]
                        }
                    }]
                }]
            }
        },
        $edit: {
            $copy: "$details"
        }
    }
}
