// syracuse-collaboration/lib/representations/roleMob.js
exports.representation = {
    $entityName: "role",
    $facets: {
        $details: {
            $layout: {
                $items: [{
                    $title: "{@InformationsSectionTitle}",
                    "$category": "section",
                    $layout: {
                        $items: [{
                            $bind: "description"
                        }, {
                            $bind: "functionProfile"
                        }]
                    }
                }, {
                    $title: "{@AdministrationSectionTitle}",
                    "$category": "section",
                    
                    $layout: {
                        $items: [{
                            $bind: "groups"
                        }]
                    }
                }]
            }
        },
        $edit: {
            $copy: "$details"
        },
        $query: {
            $layout: {
                $items: [{
                    $category: "section",
                    $layout: {
                        $items: [{
                            $bind: "$resources",
                            $format: "cards",
                            $cards: {
                                $layout: {
                                    $layoutType: "stack",
                                    $items: [{
                                        $bind: "description",
                                        $isTitleHidden: true
                                    }]
                                }
                            }
                        }]
                    }
                }]
            }
        },
        $lookup: {
            $copy: "$query"
        }
    }
}
