// syracuse-collaboration/lib/representations/role.js
exports.representation = {
    $entityName: "role",
    $facets: {
        $details: {
            $layout: {
                $items: [{
                    $title: "Informations",
                    $category: "block",
                    $layout: {
                        $items: [{
                            $bind: "description"
                        }, {
                            $bind: "functionProfile"
                        }]
                    }
                }, {
                    $title: "Administration",
                    $category: "block",
                    
                    $layout: {
                        $items: [{
                            $bind: "groups"
                        }]
                    }
                }/*, {
                    $title: "Personnalisations",
                    $category: "block",
                    $layout: {
                        $items: [{
                            $bind: "pages"
                        }]
                    }
                }*/]
            }
        },
        $edit: {
            $copy: "$details"
        }
    }
}
