// syracuse-collaboration/lib/representations/setting.js
"use strict";

var entity = require('syracuse-collaboration/lib/entities/systemInfo').entity;

exports.representation = {
    $entityName: "setting",
    $facets: {
        $details: {
            $layout: {
                $layoutType: "row",
                $items: [{
                    $items: [{
                        $title: "{@AuthenticationSectionTitle}",
                        $category: "section",
                        $layout: {
                            $items: [{
                                $bind: "authentication"
                            }, {
                                $bind: "ldap"
                            }, {
                                $bind: "oauth2"
                            }]
                        }
                    }, {
                        $items: [{
                            $title: "{@ProxySectionTitle}",
                            $category: "section",
                            $layout: {
                                $items: [{
                                    $bind: "proxy"
                                }, {
                                	$bind: "proxyConf"
                                }]
                            }
                        }]
                    }, {
                        $title: "{@PatchSectionTitle}",
                        $category: "section",
                        $layout: {
                            $items: [{
                                $bind: "patchLock"
                            }]
                        }
                    }]
                }]
            }
        },
        $edit: {
            $copy: "$details"
        }
    }
}
