// syracuse-collaboration/lib/representations/setting.js
"use strict";

exports.representation = {
	resources: function() {
		var locale = require("syracuse-core/lib/locale");
		var globals = require('streamline/lib/globals');
		return locale.resources(module, globals.context && globals.context.sessionLocale)();
	},
	$entityName: "setting",
	$facets: {
		$details: {
			$layout: {
				$layoutType: "row",
				$items: [{
					$items: [{
						$title: "{@GeneralSectionTitle}",
						$category: "section",
						$layout: {
							$items: [{
								$bind: "twoDigitYearMin"
							}, {
								$bind: "twoDigitYearMax"
							}]
						}
					}, {
						$title: "{@AuthenticationSectionTitle}",
						$category: "section",
						$layout: {
							$items: [{
								$bind: "authentication"
							}, {
								$bind: "ldap"
							}, {
								$bind: "oauth2"
							}]
						}
					}, {
						$items: [{
							$title: "{@ProxySectionTitle}",
							$category: "section",
							$layout: {
								$items: [{
									$bind: "proxy"
								}, {
									$bind: "proxyConf"
								}]
							}
						}]
					}, {
						$items: [{
							$title: "{@LoggerSectionTitle}",
							$category: "section",
							$layout: {
								$items: [{
									$bind: "logger"
								}]
							}
						}]
					}, {
						$title: "{@PatchSectionTitle}",
						$category: "section",
						$layout: {
							$items: [{
								$bind: "patchLock"
							}]
						}
					}, {
						$title: "{@SyncSectionTitle}",
						$category: "section",
						$layout: {
							$items: [{
								$bind: "conflictPriority",
							}, {
								$bind: "endpoint"
							}]
						}
					}]
				}]
			}
		},
		$edit: {
			$copy: "$details"
		}
	}
};
