// syracuse-collaboration/lib/representations/soapWebService.js
exports.representation = {
    $entityName: "soapWebService",
    $facets: {
        $details: {
            $layout: {
                $items: [{
                    $title: "{@InformationsSectionTitle}",
                    $category: "block",
                    $layout: {
                        $items: [{
                            $bind: "name"
                        }, {
                            $bind: "description"
                        }, {
                            $bind: "wsdl"
                        }]
                    }
                }, {
                    $title: "{@AdministrationSectionTitle}",
                    $category: "block",
                    $layout: {
                        $items: [{
                            $bind: "application"
                        }, {
                            $bind: "endpoint"
                        }]
                    }
                }, {
                    $title: "{@ContentSectionTitle}",
                    $category: "block",
                    $layout: {
                        $items: [{
                            $bind: "entities"
                        }]
                    }
                }]
            }
        },
        $edit: {
            $layout: {
                $items: [{
                    $title: "{@InformationsSectionTitle}",
                    $category: "block",
                    $layout: {
                        $items: [{
                            $bind: "name"
                        }, {
                            $bind: "description"
                        }]
                    }
                }, {
                    $title: "{@AdministrationSectionTitle}",
                    $category: "block",
                    $layout: {
                        $items: [{
                            $bind: "application"
                        }, {
                            $bind: "endpoint"
                        }]
                    }
                }, {
                    $title: "{@ContentSectionTitle}",
                    $category: "block",
                    $layout: {
                        $items: [{
                            $bind: "entities"
                        }]
                    }
                }]
            }
        },
        $query: {
            $layout: {
                $items: [{
                    $category: "section",
                    $layout: {
                        $items: [{
                            $bind: "$resources",

                        }]
                    }
                }]
            }
        }
    }
}
