// syracuse-collaboration/lib/representations/soapStub.js
exports.representation = {
    $entityName: "soapStub",
    $facets: {
        $details: {
            $layout: {
                $items: [{
                    $title: "{@InformationsSectionTitle}",
                    "$category": "section",
                    $layout: {
                        $items: [{
                            $bind: "name"
                        }, {
                            $bind: "url"
                        }, {
                            $bind: "description"
                        }]
                    }
                }, {
                    $title: "{@ContentSectionTitle}",
                    "$category": "section",
                    $layout: {
                        $items: [{
                            $bind: "wsdl"
                        }]
                    }
                }, {
                    $title: "{@OperationsSectionTitle}",
                    "$category": "section",
                    $layout: {
                        $items: [{
                            $bind: "ports"
                        }]
                    }
                }]
            }
        },
        $edit: {
            $layout: {
                $items: [{
                    $title: "{@InformationsSectionTitle}",
                    "$category": "section",
                    $layout: {
                        $items: [{
                            $bind: "name"
                        }, {
                            $bind: "url"
                        }, {
                            $bind: "authentication"
                        }, {
                            $bind: "authUser"
                        }, {
                            $bind: "authPassword"
                        }, {
                            $bind: "wsdl"
                        }]
                    }
                }]
            }
        },
        $query: {
            $layout: {
                $items: [{
                    $category: "section",
                    $layout: {
                        $items: [{
                            $bind: "$resources",
                            $layout: {
                                $items: [{
                                    $bind: "name"
                                }, {
                                    $bind: "url"
                                }, {
                                    $bind: "description"
                                }]
                            }

                        }]
                    }
                }]
            }
        }
    }
}
