// syracuse-collaboration/lib/representations/team.js
exports.representation = {
    $entityName: "team",
    $facets: {
        $details: {
            $layout: {
                $items: [{
                    $layoutType: "columns",
                    $layoutSubType: "50,50",
                    $items: [{
                        $items: [{
                            $title: "{@InformationsSectionTitle}",
                            $category: "block",
                            $layout: {
                                $items: [{
		                        	$bind: "description"
		                        },{
		                        	$bind: "isPublic"
		                        }]
                            }
                        },{
                            $title: "{@AdministrationSectionTitle}",
                            $category: "block",
                            $layout: {
                                $items: [{
                                	$bind: "administrator"
                                },{
                                	$bind: "authors"
                                },{
                                	$bind: "members"
                                }]
                            }
                        },{
                            $title: "{@ContentSectionTitle}",
                            $category: "block",
                            $layout: {
                                $items: [{
                                	$bind: "documents"
                                }]
                            }
                        }]
                    },{
                        $items: [{
                            $title: "{@TagsSectionTitle}",
                            $category: "block",
                            $layout: {
                                $items: [{
                                	$bind: "tags",
                                	$isTitleHidden: true
                                }]
                            }
                        }]
                    }]
                }]
            }
        },
        $edit: {
        	$copy: "$details"
        },
        $query: {
            $layout: {
                $items: [{
                    $category: "section",
                    $layout: {
                        $items: [{
                            $bind: "$resources",
                            $layout: {
	                            $items: [{
	                            	$bind: "description"
	                            },{
	                            	$bind: "isPublic"
	                            }, {
	                            	$bind: "administrator"
	                            }]
	                        }
                        }]
                    }
                }]
            }
        },
        $select: {
        	$copy: "$query"
        },
        $lookup: {
        	$copy: "$query"
        }
    }
}
