// syracuse-collaboration/lib/representations/translationTool.js
"use strict";
exports.representation = {
	$entityName: "translationTool",
	$facets: {
		$details: {
			$layout: {
				$items: [{
					$layoutType: "row",
					$items: [{
						$title: "Endpoints",
						$category: "section",
						$layout: {
							$items: [{
								$bind: "endpoint"
							}]
						}
					}]
				}, {
					$layoutType: "row",
					$items: [{
						$title: "Resource types",
						$category: "section",
						$layout: {
							$items: [{
								$bind: "syracuse"
							}, {
								$bind: "dotnet"
							}, {
								$bind: "admin"
							}]
						}
					}]
				}, {
					$layoutType: "row",
					$items: [{
						$title: "Options",
						$category: "section",
						$layout: {
							$items: [{
								$category: "section",
								$layout: {
									$items: [{
										$bind: "noDBStore"
									}, {
										$bind: "resetIndex"
									}, {
										$bind: "loadFromIndex"
									}, {
										$bind: "filterAdmin"
									}, {
										$bind: "ignoreChecksum"
									}, {
										$bind: "ignoreIndexEndpoint"
									}, {
										$bind: "verify"
									}]
								}
							}]
						}
					}]
				}]
			}
		},
		$edit: {
			$copy: "$details"
		},
		$create: {
			$copy: "$details"
		},
		$query: {
			$layout: {
				$items: [{
					$category: "section",
					$layout: {
						$items: [{
							$category: "section",
							$layout: {
								$items: [{
									$bind: "$resources"
								}]
							}
						}]
					}
				}]
			}
		}
	}
};
