// syracuse-collaboration/lib/representations/uploadVolumeItem.js
"use strict";
exports.representation = {
    $entityName: "uploadVolumeItem",
    $facets: {
        $details: {
            $layout: {
                $layoutType: "row",
                $items: [{
                    $title: "{@UploadContentSectionTitle}",
                    $category: "section",
                    $layout: {
                        $items: [{
                            $category: "section",
                            $items: [{
                                $bind: "volume"
                            }, {
                                $bind: "content"
                            }]
                        }]
                    }
                }, {
                    $title: "{@UploadMetaSectionTitle}",
                    $category: "section",
                    $layout: {
                        $items: [{
                            $category: "section",
                            $items: [{
                                $bind: "fileName"
                            }, {
                                $bind: "contentType"
                            }]
                        }]
                    }
                }]
            }
        },
        $edit: {
            $copy: "$details"
        }
    }
}
