// syracuse-collaboration/lib/representations/userMob.js
exports.representation = {
    $entityName: "user",
    $facets: {
        $details: {
            $layout: {
                $items: getPrototype(true)
            }
        },
        $edit: {
            $layout: {
                $items: getPrototype(false)
            }
        },
        $query: {
            $prototype: {
                $links: {
                    $create: {
                        $isHidden: true
                    }
                
                
                }
            },
            $layout: {
                $items: [{
                    $category: "section",
                    $layout: {
                        $items: [{
                            $bind: "$resources",
                            $format: "cards",
                            $hidePager: true,
                            $cards: {
                                $layout: {
                                    $layoutType: "side",
                                    $items: [{
                                        "$category": "section",
                                        $layoutType: "stack",
                                        $items: [{
                                            $bind: "photo",
                                            $mobile: "icon",
                                            $isTitleHidden: true
                                        }]
                                    }, {
                                        $layoutType: "stack",
                                        $items: [{
                                            $bind: "fullName",
                                            $isTitleHidden: true
                                        }, {
                                            $bind: "login",
                                            $mobile: "hint",
                                            $isTitleHidden: true
                                        }]
                                    }]
                                }
                            }
                        }]
                    }
                }]
            }
        }
    }
}


function getPrototype($isTitleHidden){
    return [{
        $layoutType: "row",
        $items: [{
            $items: [{
                $title: "{@InformationsSectionTitle}",
                "$category": "section",
                $layout: {
                    $items: [{
                        $bind: "photo",
                        $isTitleHidden: $isTitleHidden
                    }, {
                        $bind: "title",
                        $isTitleHidden: $isTitleHidden
                    }, {
                        $bind: "firstName",
                        $isTitleHidden: $isTitleHidden
                    }, {
                        $bind: "lastName",
                        $isTitleHidden: $isTitleHidden
                    }, {
                        $bind: "email",
                        $isTitleHidden: $isTitleHidden
                    }]
                }
            }, {
                $title: "{@LoginSectionTitle}",
                "$category": "section",
                $layout: {
                    $items: [{
                        $bind: "login"
                    }, {
                        $bind: "password"
                    }]
                }
            }]
        }]
    }, {
        $title: "{@AdministrationSectionTitle}",
        "$category": "section",
        $layout: {
            $items: [{
                $layoutType: "stack",
                $items: [{
                    $bind: "groups"
                }, {
                    $bind: "endpoints"
                }, {
                    $bind: "adminTeams"
                }, {
                    $bind: "authorTeams"
                }, {
                    $bind: "memberTeams"
                }]
            }]
        }
    }, {
        $title: "{@LocalesSectionTitle}",
        "$category": "section",
        $layout: {
            $items: [{
                $bind: "locales",
                $isTitleHidden: true,
                $format: "cards",
                $cards: {
                    $layout: {
                        $layoutType: "stack",
                        $items: [{
                            $bind: "code"
                        }, {
                            $bind: "description"
                        }, {
                            $bind: "enabled"
                        }, {
                            $bind: "shortDate"
                        }, {
                            $bind: "shortTime"
                        }, {
                            $bind: "shortDatetime"
                        }, {
                            $bind: "longDate"
                        }, {
                            $bind: "longTime"
                        }, {
                            $bind: "longDatetime"
                        }, {
                            $bind: "firstDayOfWeek"
                        }, {
                            $bind: "twoDigitYearMax"
                        }, {
                            $bind: "numberDecimalSeparator"
                        }, {
                            $bind: "numberGroupSeparator"
                        }, {
                            $bind: "numberGroupSize"
                        }]
                    
                    }
                }
            }]
        }
    }]
}
