// syracuse-collaboration/lib/representations/user.js
"use strict";
exports.representation = {
    $entityName: "user",
    $facets: {
        $details: {
            $layout: {
                $items: [{
                    $layoutType: "row",
                    $items: [{
                        $items: [{
                            $title: "{@LoginSectionTitle}",
                            $category: "section",
                            $layout: {
                                $items: [{
                                    $category: "section",
                                    $layout: {
                                        $items: [{
                                            $bind: "login"
                                        }, {
                                            $bind: "active",
                                        }, {
                                            $bind: "authentication",
                                        }, {
                                            $bind: "password"
                                        }, {
                                            $bind: "authenticationName"
                                        }, {
                                            $bind: "ldap"
                                        }, {
                                            $bind: "oauth2"
                                        }]
                                    }
                                }]
                            }
                        }, {
                            $title: "{@InformationsSectionTitle}",
                            $category: "section",
                            $layout: {
                                $items: [{
                                    $category: "section",
                                    $layout: {
                                        $items: [{
                                            $bind: "title"
                                        }, {
                                            $bind: "firstName"
                                        }, {
                                            $bind: "lastName"
                                        }, {
                                            $bind: "email"
                                        }, {
                                            $bind: "photo"
                                        }]
                                    }
                                }]
                            }
                        }]
                    }, {
                        $items: [{
                            $title: "{@ExplorerSectionTitle}",
                            $category: "section",
                            $layout: {
                                $items: [{
                                    $category: "section",
                                    $layout: {
                                        $items: [{
                                            $bind: "explorer",
                                            $isTitleHidden: true
                                        }]
                                    }
                                }]
                            }
                        }]
                    }]
                }, {
                    $layoutType: "stack",
                    $items: [{
                        $title: "{@AdministrationSectionTitle}",
                        $category: "section",
                        $layout: {
                            $items: [{
                                $category: "section",
                                $layout: {
                                    $layoutType: "row",
                                    $items: [{
                                        $items: [{
                                            $bind: "groups"
                                        }, {
                                            $bind: "endpoints"
                                        }]
                                    }, {
                                        $items: [{
                                            $bind: "adminTeams"
                                        }, {
                                            $bind: "authorTeams"
                                        }, {
                                            $bind: "memberTeams"
                                        }]
                                    }]
                                }
                            }]
                        }
                    }, {
                        $title: "{@LocalesSectionTitle}",
                        $category: "section",
                        $layout: {
                            $items: [{
                                $category: "section",
                                $layout: {
                                    $items: [{
                                        $bind: "locales",
                                        $isTitleHidden: true,
                                        $format: "cards",
                                        $cards: {
                                            $layout: {
                                                $layoutType: "row",
                                                $items: [{
                                                    "$layoutType": "stack",
                                                    $items: [{
                                                        $bind: "code"
                                                    }, {
                                                        $bind: "description"
                                                    }, {
                                                        $bind: "enabled"
                                                    }, {
                                                        $bind: "shortDate"
                                                    }, {
                                                        $bind: "shortTime"
                                                    }, {
                                                        $bind: "shortDatetime"
                                                    }]
                                                }, {
                                                    "$layoutType": "stack",
                                                    $items: [{
                                                        $bind: "longDate"
                                                    }, {
                                                        $bind: "longTime"
                                                    }, {
                                                        $bind: "longDatetime"
                                                    }, {
                                                        $bind: "firstDayOfWeek"
                                                    }, {
                                                        $bind: "twoDigitYearMax"
                                                    }]
                                                }, {
                                                    "$layoutType": "stack",
                                                    $items: [{
                                                        $bind: "numberDecimalSeparator"
                                                    }, {
                                                        $bind: "numberGroupSeparator"
                                                    }, {
                                                        $bind: "numberGroupSize"
                                                    }]
                                                }]
                                            }
                                        }
                                    }]
                                }
                            }]
                        }
                    }]
                
                }]
            },
			$prototype: {
				$links: {
					$create: {
						"$title": "Word report with filter",
			            "$type": "application/syracuse-word-report",
						"$url": "{$baseUrl}/users('{$uuid}')?representation=user.$details&role={$role}&reportMode=&doc_uuid={document}&format=application/syracuse-word-report",
		                "$method": "GET",
		                "$target": "blank",
	    				"$parameters": {
	    					"$url": "{$baseUrl}/msoReportModes/$template/$workingCopies?representation=msoReportMode.$edit&role={$role}",
	    					"$method": "POST",
	    					"$properties": {
	    						"dummy": {  // the user will be prompted to enter user1 value
	    							"$title": "dummy",
	    							"$type": "application/x-string"
	    						}
	    					},
	    					"msoLocaleCode": "en-USxxx",
	    					"msoCurrentRepresentation": "{$representation}" + "." + "$details"
	    				}
					}
				}
			}
        
        },
        $edit: {
            $copy: "$details"
        },
        $query: {
            $layout: {
                $items: [{
                    $category: "section",
                    $layout: {
                        $items: [{
                            $category: "section",
                            $layout: {
                                $items: [{
                                    $bind: "$resources",
                                    $layout: {
                                        $items: [{
                                            $bind: "photo"
                                        }, {
                                            $bind: "login"
                                        }, {
                                            $bind: "title"
                                        }, {
                                            $bind: "firstName"
                                        }, {
                                            $bind: "lastName"
                                        }, {
                                            $bind: "email"
                                        }, {
                                            $bind: "active"
                                        }]
                                    }
                                }]
                            }
                        }]
                    }
                }]
            }
        },
        $lookup: {
            $layout: {
                $items: [{
                    $category: "section",
                    $layout: {
                        $items: [{
                            $category: "section",
                            $layout: {
                                $items: [{
                                    $bind: "$resources",
                                    $layout: {
                                        $items: [{
                                            $bind: "photo"
                                        }, {
                                            $bind: "login"
                                        }, {
                                            $bind: "fullName"
                                        }, {
                                            $bind: "active"
                                        }, {
                                            $bind: "email"
                                        }]
                                    }
                                }]
                            }
                        }]
                    }
                }]
            }
        },
        $select: {
            $copy: "$lookup"
        }
    }
};
