// syracuse-collaboration/lib/representations/user.js
exports.representation = {
    $entityName: "user",
    $facets: {
        $details: {
            $layout: {
                $items: [{
                    $layoutType: "row",
                    $items: [{
                        $items: [{
                            $title: "{@LoginSectionTitle}",
                            $category: "section",
                            $layout: {
                                $items: [{
                                    $bind: "login"
                                }, {
                                    $bind: "active",
                                }, {
                                    $bind: "authentication",
                                }, {
                                    $bind: "password"
                                }, {
                                    $bind: "authenticationName"
                                }, {
                                    $bind: "ldap"
                                }, {
                                    $bind: "oauth2"
                                }]
                            }
                        }, {
                            $title: "{@InformationsSectionTitle}",
                            $category: "section",
                            $layout: {
                                $items: [{
                                    $bind: "title"
                                }, {
                                    $bind: "firstName"
                                }, {
                                    $bind: "lastName"
                                }, {
                                    $bind: "email"
                                }, {
                                    $bind: "photo"
                                }]
                            }
                        }]
                    }, {
                        $items: [{
                            $title: "{@ExplorerSectionTitle}",
                            $category: "section",
                            $layout: {
                                $items: [{
                                    $bind: "explorer",
                                    $isTitleHidden: true
                                }]
                            }
                        }]
                    }]
                }, {
					$layoutType: "stack",
                    $items: [{
                        $title: "{@AdministrationSectionTitle}",
                        $category: "section",
                        $layout: {
                            $layoutType: "row",
                            $items: [{
                                $items: [{
                                    $bind: "groups"
                                }, {
                                    $bind: "endpoints"
                                }]
                            }, {
                                $items: [{
                                    $bind: "adminTeams"
                                }, {
                                    $bind: "authorTeams"
                                }, {
                                    $bind: "memberTeams"
                                }]
                            }]
                        }
                    }, {
                        $title: "{@LocalesSectionTitle}",
                        $category: "section",
                        $layout: {
                            $items: [{
                                $bind: "locales",
                                $isTitleHidden: true,
                                $format: "cards",
                                $cards: {
                                    $layout: {
                                        $layoutType: "row",
                                        $items: [{
                                            "$layoutType": "stack",
                                            $items: [{
                                                $bind: "code"
                                            }, {
                                                $bind: "description"
                                            }, {
                                                $bind: "enabled"
                                            }, {
                                                $bind: "shortDate"
                                            }, {
                                                $bind: "shortTime"
                                            }, {
                                                $bind: "shortDatetime"
                                            }]
                                        }, {
                                            "$layoutType": "stack",
                                            $items: [{
                                                $bind: "longDate"
                                            }, {
                                                $bind: "longTime"
                                            }, {
                                                $bind: "longDatetime"
                                            }, {
                                                $bind: "firstDayOfWeek"
                                            }, {
                                                $bind: "twoDigitYearMax"
                                            }]
                                        }, {
                                            "$layoutType": "stack",
                                            $items: [{
                                                $bind: "numberDecimalSeparator"
                                            }, {
                                                $bind: "numberGroupSeparator"
                                            }, {
                                                $bind: "numberGroupSize"
                                            }]
                                        }]
                                    }
                                }
                            }]
                        }
                    }]
                
                }]
            }
        
        },
        $edit: {
            $copy: "$details"
        },
        $query: {
            $layout: {
                $items: [{
                    $category: "section",
                    $layout: {
                        $items: [{
                            $bind: "$resources",
                            $layout: {
                                $items: [{
                                    $bind: "login"
                                }, {
                                    $bind: "active"
                                }, {
                                    $bind: "title"
                                }, {
                                    $bind: "firstName"
                                }, {
                                    $bind: "lastName"
                                }, {
                                    $bind: "fullName"
                                }, {
                                    $bind: "email"
                                }, {
                                    $bind: "photo"
                                }]
                            }
                        }]
                    }
                }]
            }
        }
    }
}
