// syracuse-collaboration/lib/representations/x3SessionConfig.js
"use strict";

exports.representation = {
	resources: function() {
		var locale = require("syracuse-core/lib/locale");
		var globals = require('streamline/lib/globals');
		return locale.resources(module, globals.context && globals.context.sessionLocale)();
	},
	$entityName: "x3SessionConfig",
	$facets: {
		$details: {
			$layout: {
				$items: [{
					$layoutType: "row",
					$items: [{
						$items: [{
							$title: "{@RuntimeSectionTitle}",
							$category: "section",
							$layout: {
								$items: [{
									$category: "section",
									$layout: {
										$items: [{
											$bind: "runtimeLog"
										}, {
											$bind: "logFlag"
										}, {
											$bind: "directory"
										}]
									}
								}]
							}
						}, {
							$title: "{@CvgSectionTitle}",
							$category: "section",
							$layout: {
								$items: [{
									$category: "section",
									$layout: {
										$items: [{
											$bind: "cvgLog"
										}, {
											$bind: "cvgProtocol"
										}, {
											$bind: "cvgSession"
										}, {
											$bind: "cvgAction"
										}]
									}
								}]
							}
						}]
					}]
				}]
			}
		},
		$edit: {
			$copy: "$details"
		}
	}
};
