// syracuse-ui/lib/field/array/record/filter/filterCell.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function FilterCell(){
}

exports.FilterCell = helpers.defineClass(FilterCell, null, {
    addOperator: function(field){
        var self = this;
        if (field._input) {
            field._input.className = field._input.className.replace("s-inplace-input", "s-filter-input");
        }
        self._operator = document.createElement("a");
        self._operator.className = "s-list-filter";
        self._operator.style.backgroundImage = "url('" + document.site.$item.$iconPath + "filter/s-filter-none.png')";
        $(field._dataValue.appendChild(self._operator)).bind("click", function(){
            if (!self.listPopup) {
                self.$$list.empty();
                var operators;
                switch (self.$field.$type) {
                    case "application/x-date":
                    case "application/x-time":
                    case "application/x-datetime":
                        operators = ["none", "gt", "ge", "lt", "le", "eq", "ne", "between"];
                        break;
                    case "application/x-string":
                        operators = ["none", "gt", "ge", "lt", "le", "eq", "ne", "like_s", "like", "between"];
                        break;
                    default:
                        operators = ["none", "eq", "ne"];
                        break;
                }
                operators.forEach(function(op){
                    var link = document.createElement("a");
                    link.className = "s-list-filter-item";
                    link.setAttribute("data-s-op", op)
                    link.style.backgroundImage = "url('" + document.site.$item.$iconPath + "filter/s-filter-" + op + ".png')";
                    $(link).appendTo(self.$$list).text(self.record.boxParent._localize["filter_" + op]);
                });
                self.listPopup = self.record.boxParent.openDialog({
                    $dialogMode: "popup",
                    content: self.record,
                    $$dialog: self.$$list,
                    position: {
                        my: "left top",
                        at: "left bottom",
                        of: self.$$container
                    },
                    onClose: function(){
                        self.listPopup = null;
                    }
                });
            }
            else {
                self.listPopup.close();
            }
            return false;
        });
        if (!self.$$list) {
            var list = document.createElement("div");
            list.className = "s-list-filter-list";
            list.style.display = "none";
            self.$$list = $(field._dataValue.appendChild(list)).delegate("a.s-list-filter-item", "click", function(){
                self.setOperator($(this).attr("data-s-op"));
                self.listPopup.close()
                // between, create the "high" field
                var highField = self.record.boundFields[self.$bind + "$High"];
                if (self.operator === "between") {
                    if (highField) {
                        highField[0].$$dataValue.toggle(true);
                    }
                    else {
                        self.loadField(self.$field, self.$bind + "$High");
                    }
                }
                else {
                    if (highField) {
                        highField[0].$$dataValue.toggle(false);
                    }
                    var value = self.record.dataset[self.$bind];
                    if (self.operator != "none") {
                        if (value != null && value != "") {
                            self.record._buildFilter();
                        }
                    }
                    else {
                        if (value != null && value != "") {
                            var field = self.record.boundFields[self.$bind][0];
                            field.setDataValue(null);
                            field.notifyFieldChange(null);
                        }
                        else {
                            self.record._buildFilter();
                        }
                    }
                }
            });
        }
    },
    render: function(record, $$container, $field, $bind){
        this.$field = $field;
        this.record = record;
        this.$bind = $bind;
        this.operator = "none";
        this.$$container = $$container;
        this.loadField(self.$field, self.$bind);
    },
    setOperator: function(operator){
        this._operator.style.backgroundImage = "url('" + document.site.$item.$iconPath + "filter/s-filter-" + (this.operator = operator) + ".png')";
    },
    _format: function(field, $type, value){
        switch ($type) {
            case "application/x-reference":
                return "'" + (field.currentValue.$key || field.currentValue.$value) + "'";
            case "application/x-choice":
                var fieldType = (field.$field && field.$field.$value && field.$field.$value.$type) || "application/x-string";
                return this._format(field, fieldType, value);
            case "application/x-boolean":
            case "application/x-integer":
            case "application/x-decimal":
                return value;
            case "application/x-date":
                return "@" + value + "@";
        }
        return "'" + value + "'";
    },
    getValue: function(){
        var field = this.record.boundFields[this.$bind][0];
        var value = this.record.dataset[this.$bind];
        if (this.operator != "none" && value != null) {
            var op = this.operator;
            var right;
            switch (op) {
                case "like":
                    // like is for strings only so direct access to store
                    right = "'%" + value + "%'";
                    break;
                case "like_s":
                    // like is for strings only so direct access to store
                    right = "'" + value + "%'";
                    op = "like";
                    break;
                case "between":
                    var highValue = this.record.dataset[this.$bind + "$High"];
                    if (value && !highValue) {
                        right = this._format(field, this.$field.$type, value);
                        op = "ge";
                    }
                    else {
                        if (!value && highValue) {
                            right = this._format(field, this.$field.$type, highValue);
                            op = "le";
                        }
                        else {
                            if (value && highValue) {
                                right = this._format(field, this.$field.$type, value) + " and " + this._format(field, this.$field.$type, highValue);
                            }
                            else {
                                return "";
                            }
                        }
                    }
                    break;
                default:
                    right = this._format(field, this.$field.$type, value);
                    break;
            }
            return "(" + this.$bind + " " + op + " " + right + ")";
        }
    },
    loadField: function($field, $bind){
        this.addOperator(document.itemFactory.load(this.$$container, {
            $bind: $bind || this.$bind,
            $isTopLabelAlignment: false,
            $isCellChild: true,
            $isFilterMode: true,
            $inplace: true,
            $field: $field,
            $isBindInputEvent: true
        }, this.record));
    },
    ensureEqual: function(){
        // force operator to equals
        if (this.operator === "none" && this.$field) {
            var defaultFilter;
            switch (this.$field.$type) {
                case "application/x-date":
                case "application/x-time":
                case "application/x-datetime":
                    defaultFilter = "ge";
                    break;
                case "application/x-string":
                    defaultFilter = "like";
                    break;
                default:
                    defaultFilter = "eq";
                    break;
            }
            this.setOperator(defaultFilter);
        }
    },
    dispose: function(){
        if (this._operator) {
            $(this._operator).unbind();
        }
        if (this.$$list) {
            this.$$list.undelegate();
        }
        delete this.record;
    },
    // FDB 
    fusionGetInputValue: function(){
        var field = this.record.boundFields[this.$bind][0];
        return field.getDataValue();
    }
});
