// syracuse-ui/lib/site/search/searchController.js
"use strict";

exports.isSearchUrl = function($url){
    return $url && ($url.indexOf("$search?representation=queryResult.$search&q=") >= 0);
};
exports.openRequestPage = function(httpQuery){
    //    document.controller.loadRepresentation(null, httpQuery, function($ip){
    //        var searchQuery = document.controller.parseUrl(httpQuery.$url);
    //        var $itemPage = {
    //            $category: "search",
    //            $urlParts: searchQuery.$urlParts,
    //            $representation: $ip.$representation
    //        };
    //        $itemPage.$urlParts.$facet = "$query";
    //        $itemPage.httpQuery = httpQuery;
    //        document.site.onMainPageChange($itemPage);
    //    });
    //var searchQuery = document.controller.parseUrl(httpQuery.$url);
    //searchQuery.sendRequest({}, function(data, response, requestUrl){
    
    //});
    // TODO: onerror functions
    
    var $itemPage = {
        $category: "search",
        $urlParts: httpQuery.$urlParts,
        $representation: {
            $prototype: {
                "$baseUrl": "http://localhost:8126/sdata/syracuse/search/syracuse",
                "$baseType": "application/json;vnd.sage=syracuse",
                "$url": "{$baseUrl}/queryResults",
                "$type": "application/json",
                "$title": "Search results",
                "$properties": {
                    "$queryFunctions": {
                        "$type": "application/x-array",
                        "$item": {
                            "$type": "application/json",
                            "$prototype": "{$baseUrl}/$prototypes('{$representation}.$search')",
                            "$url": "{$baseUrl}/{$pluralType}('{$key}')",
                            "$value": "{$key}",
                            "$title": "queryResult {$key}",
                            "$key": "{$uuid}",
                            "$description": "queryResult {$key}",
                            "$pluralType": "queryResults",
                            "$representation": "queryResult",
                            "$properties": {
                                "type": {
                                    "$title": "Type",
                                    "$type": "application/x-string",
                                    "$capabilities": "sort,filter,alphaTab"
                                },
                                "description": {
                                    "$title": "Description",
                                    "$type": "application/x-string",
                                    "$capabilities": "sort,filter,alphaTab",
                                    "$links": {
                                        "$details": {
                                            "$type": "application/json;vnd.sage=syracuse",
                                            "$url": "{$baseUrl}/{$pluralType}('{$key}')?representation={$representation}.$details&role={$role}"
                                        }
                                    }
                                },
                                "$creUser": {
                                    "$title": "Created by",
                                    "$type": "application/x-string",
                                    "$isDisabled": true,
                                    "$isHidden": true,
                                    "$capabilities": "sort,filter"
                                },
                                "$updUser": {
                                    "$title": "Updated by",
                                    "$type": "application/x-string",
                                    "$isDisabled": true,
                                    "$isHidden": true,
                                    "$capabilities": "sort,filter"
                                },
                                "$creDate": {
                                    "$title": "Created on",
                                    "$type": "application/x-datetime",
                                    "$isDisabled": true,
                                    "$isHidden": true,
                                    "$capabilities": "sort,filter"
                                },
                                "$updDate": {
                                    "$title": "Updated on",
                                    "$type": "application/x-datetime",
                                    "$isDisabled": true,
                                    "$isHidden": true,
                                    "$capabilities": "sort,filter"
                                }
                            },
                            "$links": {
                                "$details": {
                                    "$type": "application/json;vnd.sage=syracuse",
                                    "$title": "Detail",
                                    "$url": "{$baseUrl}/{$pluralType}('{$key}')?representation={$representation}.$details&role={$role}"
                                },
                                "$edit": {
                                    "$type": "application/json;vnd.sage=syracuse",
                                    "$title": "Edit",
                                    "$url": "{$baseUrl}/{$pluralType}('{$key}')/$workingCopies?representation={$representation}.$edit&role={$role}",
                                    "$method": "POST"
                                },
                                "$delete": {
                                    "$title": "Delete",
                                    "$confirm": "The record undefined will be deleted. Confirm ?",
                                    "$url": "{$baseUrl}/{$pluralType}('{$key}')?representation={$representation}.$search&role={$role}",
                                    "$type": "application/json;vnd.sage=syracuse",
                                    "$method": "DELETE"
                                },
                                "$query": {
                                    "$type": "application/json;vnd.sage=syracuse",
                                    "$title": "Search results",
                                    "$url": "{$baseUrl}/{$pluralType}?representation={$representation}.$query&role={$role}"
                                },
                                "$print": {
                                    "$type": "application/pdf",
                                    "$title": "Print"
                                }
                            },
                            "$defaultOrder": "description"
                        }
                    },
                    "$queryData": {
                        "$type": "application/x-array",
                        "$item": {
                            "$type": "application/json",
                            "$prototype": "{$baseUrl}/$prototypes('{$representation}.$search')",
                            "$url": "{$baseUrl}/{$pluralType}('{$key}')",
                            "$value": "{$key}",
                            "$title": "queryResult {$key}",
                            "$key": "{$uuid}",
                            "$description": "queryResult {$key}",
                            "$pluralType": "queryResults",
                            "$representation": "queryResult",
                            "$properties": {
                                "type": {
                                    "$title": "Type",
                                    "$type": "application/x-string",
                                    "$capabilities": "sort,filter,alphaTab"
                                },
                                "description": {
                                    "$title": "Description",
                                    "$type": "application/x-string",
                                    "$capabilities": "sort,filter,alphaTab",
                                    "$links": {
                                        "$details": {
                                            "$type": "application/json;vnd.sage=syracuse",
                                            "$url": "{$baseUrl}/{$pluralType}('{$key}')?representation={$representation}.$details&role={$role}"
                                        }
                                    }
                                },
                                "$creUser": {
                                    "$title": "Created by",
                                    "$type": "application/x-string",
                                    "$isDisabled": true,
                                    "$isHidden": true,
                                    "$capabilities": "sort,filter"
                                },
                                "$updUser": {
                                    "$title": "Updated by",
                                    "$type": "application/x-string",
                                    "$isDisabled": true,
                                    "$isHidden": true,
                                    "$capabilities": "sort,filter"
                                },
                                "$creDate": {
                                    "$title": "Created on",
                                    "$type": "application/x-datetime",
                                    "$isDisabled": true,
                                    "$isHidden": true,
                                    "$capabilities": "sort,filter"
                                },
                                "$updDate": {
                                    "$title": "Updated on",
                                    "$type": "application/x-datetime",
                                    "$isDisabled": true,
                                    "$isHidden": true,
                                    "$capabilities": "sort,filter"
                                }
                            },
                            "$links": {
                                "$details": {
                                    "$type": "application/json;vnd.sage=syracuse",
                                    "$title": "Detail",
                                    "$url": "{$baseUrl}/{$pluralType}('{$key}')?representation={$representation}.$details&role={$role}"
                                },
                                "$edit": {
                                    "$type": "application/json;vnd.sage=syracuse",
                                    "$title": "Edit",
                                    "$url": "{$baseUrl}/{$pluralType}('{$key}')/$workingCopies?representation={$representation}.$edit&role={$role}",
                                    "$method": "POST"
                                },
                                "$delete": {
                                    "$title": "Delete",
                                    "$confirm": "The record undefined will be deleted. Confirm ?",
                                    "$url": "{$baseUrl}/{$pluralType}('{$key}')?representation={$representation}.$search&role={$role}",
                                    "$type": "application/json;vnd.sage=syracuse",
                                    "$method": "DELETE"
                                },
                                "$query": {
                                    "$type": "application/json;vnd.sage=syracuse",
                                    "$title": "Search results",
                                    "$url": "{$baseUrl}/{$pluralType}?representation={$representation}.$query&role={$role}"
                                },
                                "$print": {
                                    "$type": "application/pdf",
                                    "$title": "Print"
                                }
                            },
                            "$defaultOrder": "description"
                        }
                    }
                },
                "$links": {
                    "$create": {
                        "$type": "application/json;vnd.sage=syracuse",
                        "$title": "Create",
                        "$url": "{$baseUrl}/{$pluralType}/$template/$workingCopies?representation={$representation}.$edit&role={$role}",
                        "$method": "POST"
                    },
                    "$print": {
                        "$type": "application/pdf",
                        "$title": "Print"
                    },
                    "$excel": {
                        "$type": "application/syracuse-excel-worksheet",
                        "$title": "Excel"
                    }
                },
                "$descriptor": "prototype queryResult.$search",
                "$article": {
                    "$layout": {
                        "$layoutType": "tabs",
                        "$items": [{
                            "$category": "section",
                            "$title": "Functions",
                            "$layout": {
                                "$layoutType": "columns",
                                "$layoutSubType": "50,50",
                                "$items": [{
                                    "$items": [{
                                        "$bind": "$queryFunctions",
                                        "$skin": "toto",
                                        "$format": "cards",
                                        "$cards": {
                                            "$layout": {
                                                "$items": [{
                                                    "$bind": "description",
                                                    "$isTitleHidden": true,
                                                    "$css": "tutu"
                                                }, {
                                                    "$bind": "type",
                                                    "$isTitleHidden": true,
                                                    "$skin": "popi"
                                                }]
                                            }
                                        }
                                    }]
                                }]
                            }
                        }, {
                            "$category": "section",
                            "$title": "Data",
                            "$layout": {
                                "$layoutType": "columns",
                                "$layoutSubType": "50,50",
                                "$items": [{
                                    "$items": [{
                                        "$bind": "$queryData",
                                        "$format": "cards",
                                        "$cards": {
                                            "$layout": {
                                                "$items": [{
                                                    "$bind": "description",
                                                    "$isTitleHidden": true
                                                }, {
                                                    "$bind": "type",
                                                    "$isTitleHidden": true
                                                }]
                                            }
                                        }
                                    }]
                                }]
                            }
                        }]
                    }
                }
            }
        }
    }
    
    document.site.onMainPageChange($itemPage);
    
};
