// syracuse-ui/lib/site/search/searchField.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Field = require('syracuse-ui/lib/field/field').Field;

function SearchField() {}

exports.SearchField = helpers.defineClass(SearchField, Field, {
	onInputChange: function(input, event) {
		// this.onInputValidate(event);
	},
	onClickPicker: function(btn) {
		// handle clear button behaviour
		if (btn && btn.syraPickerType == "clear") {
			this.setDataValue("");
			btn.style.display = "none";
		}
		this.onInputValidate(); //trigger notification change
	},
	onInputFocusin: function(input, event) {
		// handle clear button behaviour
		if (this.getDataValue() != "") {
			this.picker.style.display = "inline-block";
			this.picker.style.opacity = 1;
		}
		//  Field.prototype.onInputFocusin.call(this, input, event);
	},
	onInputFocusout: function(input, event) {
		// handle clear button behaviour
		if (this.getDataValue() != "") {
			this.picker.style.display = "inline-block";
			this.picker.style.opacity = 0.4;
		}
		//Field.prototype.onInputFocusout.call(this, input, event);
	},
	onKeyUp: function(input, event) {
		this.picker.style.display = this.getDataValue() == "" ? "none" : "inline-block";
		this.picker.style.opacity = 1;
	},
	applyShortCuts: function(shortcurts) {
		var self = this;
		if (this._isAutoComplete) {
			// don't run search if 'Enter' button
			if (!shortcurts.enter) {
				//stop any previous ongoing search request
				clearTimeout(this._searching);
				// run search
				this.runAutoCompleteSearch();
				if (this._autoCompletePopup) { // handle navigation in drop down search result list (down(40) and up(38))
					if ((shortcurts.up || shortcurts.down) && this._autoCompletePopup) {
						var searchResRecords = this.page.boundFields.$query1[0].records.concat(this.page.boundFields.$query2[0].records);
						var index = this._getFocusRecIndex(searchResRecords);
						var jj = searchResRecords.length;
						if (index != undefined) {
							searchResRecords[shortcurts.down ? ((index + 1) <= (jj - 1) ? (index + 1) : 0) : ((index - 1) >= 0 ? (index - 1) : jj - 1)].$$item.addClass("s-search-rq-field-carditem-hover");
						} else {
							searchResRecords[shortcurts.down ? 0 : jj - 1].$$item.addClass("s-search-rq-field-carditem-hover");
						}
						this.$isArrowClicked = true;
					}
				}
			} else {
				// remove autocomplete list if opened
				if (this._autoCompletePopup) {
					// handle navigation in drop down search result list (enter)
					var searchResRecords = this.page.boundFields.$query1[0].records.concat(this.page.boundFields.$query2[0].records);
					var index = this._getFocusRecIndex(searchResRecords);
					if (index != undefined && this.$isArrowClicked) {
						var record = searchResRecords[index];
						record.$$item.click();
						this.$isRecordTriggered = true;
					} else {
						// run search if field not empty
						if (this.getDataValue() !== '') {
							this.onInputValidate();
						}
						// close popup
						this._autoCompletePopup.close();
					}
				} else {
					if (this.getDataValue() !== '') {
						clearTimeout(this._searching);
						clearTimeout(this._onFieldChangeNotification);
						this._onFieldChangeNotification = setTimeout(function() {
							self.notifyFieldChange(self.getDataValue());
						}, 200);
					}
				}
			}
		}
	},
	runAutoCompleteSearch: function(event) {
		var self = this;
		self._searching = setTimeout(function() {
			// search doesn't take whitespace into account
			if (self.getDataValue().replace(/\s+/g, '') !== "") {

				// if some research has already been done, proceed with the request if the value has changed
				if (self.searchValue) {

					if (self.searchValue !== self.getDataValue()) {
						self.toggleAutoCompleteList(true);
					} else {
						// if empty field, close incremental search popup
						if (self.searchValue == self.getDataValue() == "") {
							if (self._autoCompletePopup) {
								self._autoCompletePopup.close();
							}
						}
					}
				} else {
					// first research (self.searchValue not set yet)
					self.toggleAutoCompleteList(true);
				}
			} else {
				if (self._autoCompletePopup) {
					self._autoCompletePopup.close();
				}
			}

		}, 1000);
	},
	_onAutoCompleteOpen: function() {
		return this.page.loadAutoCompleteList(this);
	},
	_onAutoCompleteClose: function() {
		this.page.unloadAutoCompleteList();
	},
	toggleAutoCompleteList: function(show) {
		var self = this;

		if (self.getDataValue() !== "") {
			self.domItem.id = self.id;
			self.focus();

			self.searchValue = self.getDataValue();

			if (self._autoCompletePopup) {
				self._onAutoCompleteOpen();
			} else {
				self._autoCompletePopup = document.site.dialogManager.openPopup(self.boxParent, {
					content: self,
					slot: self._onAutoCompleteOpen(),
					position: {
						my: "right top",
						at: "right bottom",
						of: self.$$input
					},
					onClose: function() {
						self.searchValue = null;
						self._autoCompletePopup = null;
						self._onAutoCompleteClose();
					}
				});
			}
		} else {
			if (self._autoCompletePopup) {
				self._autoCompletePopup.close();
			}
		}
	},
	filterMenu: function($bind) {
		if ($bind == "$suggest") {
			this._isAutoComplete = true;
			return false;
		}
		return Field.prototype.filterMenu.call(this, $bind);
	},
	_appendLinksPicker: function() {
		if (this.$rootLinks.$links && this.$rootLinks.$links.$suggest) {
			this._isAutoComplete = true;
			this.picker = this._appendSpecialPicker("clear");
		}
		return Field.prototype._appendLinksPicker.call(this);
	},
	_getFocusRecIndex: function(records) {
		var index;
		for (var ii = 0, jj = records.length; ii < jj; ii++) {
			var rec = records[ii];
			index = rec.$$item.hasClass("s-search-rq-field-carditem-hover") ? ii : index;
			if (index != undefined) {
				// remove focus class
				rec.$$item.removeClass("s-search-rq-field-carditem-hover");
				break;
			}
		}
		return index;
	},
	dispose: function() {
		this.picker = this.$isArrowClicked = this.$isRecordTriggered = null;
		Field.prototype.dispose.call(this);
	}
});
