// syracuse-ui/lib/site/search/searchPage.js
"use strict";
var helpers = require("syracuse-core/lib/helpers")
var DesktopPage = require("syracuse-ui/lib/article/page").DesktopPage;

function SearchPage(){
}

exports.SearchPage = helpers.defineClass(SearchPage, DesktopPage, {
    fetch: function(){
        var self = this;
        
        var $queryFunctionsLink = {
            $url: "/sdata/syracuse/search/syracuse/$search?representation=queryResult.$search&q={$search}&researchType='functions'"
        };
        var $queryDataLink = {
            $url: "/sdata/syracuse/search/syracuse/$search?representation=queryResult.$search&q={$search}&researchType='data'"
        }
        self.dataset.$search = self.$urlParts.params.q;
        self.formatMenuUrl($queryFunctionsLink, self.dataset, function($url, isCanceled){
            document.controller.sendRequest(self, {
                $location: {
                    $url: $url
                }
            }, function(data){
                data.$itemsPerPage = 5;
                data.$queryFunctions = data.$resources;
                delete data.$resources;
                self.applyChange(data);
            });
        });
        self.formatMenuUrl($queryDataLink, self.dataset, function($url, isCanceled){
            document.controller.sendRequest(self, {
                $location: {
                    $url: $url
                }
            }, function(data){
                data.$itemsPerPage = 5;
                data.$queryData = data.$resources;
                delete data.$resources;
                self.applyChange(data);
            });
        });
    }
	/*,
    _ensureMenus: function(){
        if (this.$pageCategory == "vignette") {
            if (this.$facet != "$menu") {
                RawPage.prototype.appendArticleMenus.call(this);
            }
        }
        else {
            this.menuBar = new MenuBar();
            this.menuBar.load(this);
        }
        if (this.menuBar && this.menuBar._bar && this.$isSearchFacetEnabled) {
            this.loadNewItem($(this.menuBar._bar), {
                $category: "searchFacets",
                $prototype: {},
                $bind: "$searchFacets"
            });
        }
    }*/
});
