// syracuse-ui/lib/site/search/searchController.js
"use strict";

exports.isSearchUrl = function($url){
    return $url && ($url.indexOf("$search?representation=queryResult.$search&q=") >= 0);
};
exports.openRequestPage = function(httpQuery){
    var $itemPage = {
        $category: "search",
        $urlParts: httpQuery.$urlParts,
        $representation: {
            $prototype: {
                "$baseUrl": "http://localhost:8126/sdata/syracuse/search/syracuse",
                "$baseType": "application/json;vnd.sage=syracuse",
                "$url": "{$baseUrl}/queryResults",
                "$type": "application/json",
                "$title": "Search results",
                "$properties": {
                    "$queryFunctions": {
                        "$type": "application/x-array",
                        "$item": {
                            "$type": "application/json",
                            "$prototype": "{$baseUrl}/$prototypes('{$representation}.$search')",
                            "$url": "{$baseUrl}/{$pluralType}('{$key}')",
                            "$value": "{$key}",
                            "$title": "queryResult {$key}",
                            "$key": "{$uuid}",
                            "$description": "queryResult {$key}",
                            "$pluralType": "queryResults",
                            "$representation": "queryResult",
                            "$properties": {
                                "type": {
                                    "$title": "Type",
                                    "$type": "application/x-string",
                                    "$capabilities": "sort,filter,alphaTab"
                                },
                                "description": {
                                    "$title": "Description",
                                    "$type": "application/x-string",
                                    "$capabilities": "sort,filter,alphaTab",
                                    "$links": {
                                        "$details": {
                                            "$type": "application/json;vnd.sage=syracuse",
                                            "$url": "{$baseUrl}/{$pluralType}('{$key}')?representation={$representation}.$details&role={$role}"
                                        }
                                    }
                                },
                                "$creUser": {
                                    "$title": "Created by",
                                    "$type": "application/x-string",
                                    "$isDisabled": true,
                                    "$isHidden": true,
                                    "$capabilities": "sort,filter"
                                },
                                "$updUser": {
                                    "$title": "Updated by",
                                    "$type": "application/x-string",
                                    "$isDisabled": true,
                                    "$isHidden": true,
                                    "$capabilities": "sort,filter"
                                },
                                "$creDate": {
                                    "$title": "Created on",
                                    "$type": "application/x-datetime",
                                    "$isDisabled": true,
                                    "$isHidden": true,
                                    "$capabilities": "sort,filter"
                                },
                                "$updDate": {
                                    "$title": "Updated on",
                                    "$type": "application/x-datetime",
                                    "$isDisabled": true,
                                    "$isHidden": true,
                                    "$capabilities": "sort,filter"
                                }
                            },
                            "$defaultOrder": "description"
                        }
                    },
                    "$queryData": {
                        "$type": "application/x-array",
                        "$item": {
                            "$type": "application/json",
                            "$prototype": "{$baseUrl}/$prototypes('{$representation}.$search')",
                            "$url": "{$baseUrl}/{$pluralType}('{$key}')",
                            "$value": "{$key}",
                            "$title": "queryResult {$key}",
                            "$key": "{$uuid}",
                            "$description": "queryResult {$key}",
                            "$pluralType": "queryResults",
                            "$representation": "queryResult",
                            "$properties": {
                                "type": {
                                    "$title": "Type",
                                    "$type": "application/x-string",
                                    "$capabilities": "sort,filter,alphaTab"
                                },
                                "description": {
                                    "$title": "Description",
                                    "$type": "application/x-string",
                                    "$capabilities": "sort,filter,alphaTab",
                                    "$links": {
                                        "$details": {
                                            "$type": "application/json;vnd.sage=syracuse",
                                            "$url": "{$baseUrl}/{$pluralType}('{$key}')?representation={$representation}.$details&role={$role}"
                                        }
                                    }
                                },
                                "$creUser": {
                                    "$title": "Created by",
                                    "$type": "application/x-string",
                                    "$isDisabled": true,
                                    "$isHidden": true,
                                    "$capabilities": "sort,filter"
                                },
                                "$updUser": {
                                    "$title": "Updated by",
                                    "$type": "application/x-string",
                                    "$isDisabled": true,
                                    "$isHidden": true,
                                    "$capabilities": "sort,filter"
                                },
                                "$creDate": {
                                    "$title": "Created on",
                                    "$type": "application/x-datetime",
                                    "$isDisabled": true,
                                    "$isHidden": true,
                                    "$capabilities": "sort,filter"
                                },
                                "$updDate": {
                                    "$title": "Updated on",
                                    "$type": "application/x-datetime",
                                    "$isDisabled": true,
                                    "$isHidden": true,
                                    "$capabilities": "sort,filter"
                                }
                            },
                            "$defaultOrder": "description"
                        }
                    }
                },
                "$links": {
                    "$create": {
                        "$type": "application/json;vnd.sage=syracuse",
                        "$title": "Create",
                        "$url": "{$baseUrl}/{$pluralType}/$template/$workingCopies?representation={$representation}.$edit&role={$role}",
                        "$method": "POST"
                    },
                    "$print": {
                        "$type": "application/pdf",
                        "$title": "Print"
                    },
                    "$excel": {
                        "$type": "application/syracuse-excel-worksheet",
                        "$title": "Excel"
                    }
                },
                "$descriptor": "prototype queryResult.$search",
                "$article": {
                    "$layout": {
                        "$layoutType": "tabs",
                        "$items": [{
                            "$category": "section",
                            "$title": "Functions",
                            "$layout": {
                                "$layoutType": "columns",
                                "$layoutSubType": "50,50",
                                "$items": [{
                                    "$items": [{
                                        "$bind": "$queryFunctions",
                                        "$format": "cards",
                                        "$cards": {
                                            "$layout": {
                                                "$items": [{
                                                    "$bind": "description",
                                                    "$isTitleHidden": true
                                                }, {
                                                    "$bind": "type",
                                                    "$isTitleHidden": true
                                                }]
                                            }
                                        }
                                    }]
                                }]
                            }
                        }, {
                            "$category": "section",
                            "$title": "Data",
                            "$layout": {
                                "$layoutType": "columns",
                                "$layoutSubType": "50,50",
                                "$items": [{
                                    "$items": [{
                                        "$bind": "$queryData",
                                        "$format": "cards",
                                        "$cards": {
                                            "$layout": {
                                                "$items": [{
                                                    "$bind": "description",
                                                    "$isTitleHidden": true
                                                }, {
                                                    "$bind": "type",
                                                    "$isTitleHidden": true
                                                }]
                                            }
                                        }
                                    }]
                                }]
                            }
                        }]
                    }
                }
            }
        }
    }
    
    document.site.onMainPageChange($itemPage);
    
};
