// @sage/syracuse-ui/lib/site/config.js
"use strict";
exports.$legalDataUrl = "/sdata/syracuse/collaboration/syracuse/licenseDatas?representation=licenseData.$query";
exports.$legalPrototypeUrl = "/sdata/syracuse/collaboration/syracuse/$prototypes('licenseData.$query')";
exports.$eulaUrl = "/syracuse-ui/lib/help/SAGE X3 EULA.pdf";
exports.$eulaDepsUrl = "/syracuse-main/html/main.html?url=/sdata/syracuse/collaboration/syracuse/copyrights?representation=copyright.$query&count&orderBy=name ASC";

exports.$landingUrl = "/sdata/syracuse/collaboration/syracuse/landingVignetteSelects/$template/$workingCopies?representation=landingVignetteSelect.$edit";
exports.$searchUrl = "/sdata/syracuse/search/syracuse/$search?representation=queryResult.$search";

exports.$iconPath = "/syracuse-main/html/styles/images/";
exports.$pageTemplateUrl = "/sdata/syracuse/collaboration/syracuse/pages('{endpoint}.{representation},{target},{pageview}')?role={$role}&profile={$userprofile}";

exports.$sitemapUrl = "?representation=home.$navigation&profile={$userprofile}";
exports.$help_designer = "/help/{$localeCode}/tutorial/central-page";
exports.$support = "http://www.sage.com/support";

exports.size = {
	minWidth: 650,
	overResizeMinWidth: 400,
	overResizeMinHeight: 300,
	gridRowHeight: 28,
	textHeight: 32
};
