// syracuse-ui/lib/site/widget/extendFactory.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Field = require('syracuse-ui/lib/field/field').Field;

var Widget = helpers.defineClass(function() {}, Field, {
	loadBox: function() {
		this.$item.$isTitleHidden = true;
		Field.prototype.loadBox.call(this);
	},
	_createValueContainer: function() {
		var css = this._$cssField + (this.$isEditMode ? "-value-edit" : "-value-read");
		this.domItem.className += " " + this._$cssField;
		this._dataValue = this._core = this.domItem;
	},
	render: function() {
		var self = this;
		var editable = this.page.$facet === '$edit';
		this.component = this.api.create({
			div: this.domItem,
			facet: this.page.$facet,
			editable: editable,
			setDirty: function() {
				if (editable) self.page.applyChange({
					$actions: {
						$save: {
							$isDisabled: false,
						}
					}
				});
			}
		});
	},
	validateInput: function(errors, value) {
		//if (value && !value.isValid) {
		//	errors.push("Cannot save: uml contains errors");
		//}
	},
	getDataValue: function() {
		return this.component && this.component.getValue && this.component.getValue();
	},
	setDataValue: function(value) {
		if (this.component && this.component.setValue) this.component.setValue(value);
	},
});

exports.makeClass = function(type, api) {
	return helpers.defineClass(function() {
		Widget.call(this);
	}, Widget, {
		api: api
	});
};
