// syracuse-ui/lib/desktop/field/collection/list.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var localize = require('syracuse-ui/lib/common/controller/localize').list;
var Pager = require("./pager").Pager;
var RawList = require("syracuse-ui/lib/common/field/collection/list").RawList;
var authoringOpener = require("./authoring/authoringOpener");
var RecordReorder = require('./recordReorder').RecordReorder;

var GraphDecorator = require("./graphDecorator").GraphDecorator;

var gridBuilder = require("./builder/hgridBuilder").Builder;

var _formatBuilders = {
    array: require("./builder/arrayBuilder").Builder,
    reference: require("./builder/referenceListBuilder").Builder,
    grid: gridBuilder,
    vgrid: gridBuilder,
    cardList: require("./builder/cardListBuilder").Builder,
};


function _doSpeedTest($speedTest, $resources){
    if ($resources) {
        var $mems = [];
        for (var ii = 0; ii < self.$item.$speedTest; ii++) {
            $resources.forEach(function(record, $recordIndex){
                record.$uuid = document.controller.generateUUID();
                $mems.push(record);
            });
        }
        $resources = $mems;
        alert("start display");
    }
}


function DesktopList(){
}

exports.DesktopList = helpers.defineClass(DesktopList, RawList, {
    _initilializeList: function(){
        this.$isLayoutContentSizeDisabled = true;
        RawList.prototype._initilializeList.call(this);
        if (this.$prototype.$cube) {
            if ((this.$item.$graphMode = this.$item.$graphMode || "both") == "both") {
                this.$item.$graphPosition = "top";
            }
        }
    },
    createBuilder: function(){
        if (this.$field.$item.$type == "application/x-reference") {
            //debugger;
        }
        if (this.$field.$item.$type == "application/x-reference" && !this.$field.$newControl) {
            this.builder = new _formatBuilders.reference();
        }
        else {
            if (!this.$field.$item.$ || this.$field.$item.$type == "application/x-reference") {
                this.builder = new _formatBuilders.array();
            }
            else {
                this.$item.$format = this.$item.$format || "grid";
                this.builder = new _formatBuilders[this.$item.$format]();
                this.builder.$layoutOptions = this.$layoutOptions.builder[this.$item.$format] || null;
            }
        }
        this.builder.list = this;
        this.builder.initialize();
    },
    reloadBuilder: function(){
        var self = this;
        setTimeout(function(){
            document.controller.disposeObject(self.builder);
            $(self._core).empty();
            self.createBuilder();
            self.builder.drawBuilder();
        }, 20);
    },
    applyDesignMetaData: function(metadata, onAuthoring){
        if (onAuthoring) {
            if (metadata.$format !== undefined) {
                this.$item.$format = metadata.$format;
                this.reloadBuilder();
            }
            if (metadata.$isQuickFilter !== undefined) {
                this.$item.$isQuickFilter = metadata.$isQuickFilter;
                this.reloadBuilder();
            }
            if (metadata.$itemsPerPage !== undefined) {
                this.$item.$itemsPerPage = metadata.$itemsPerPage;
                this.onNotifyRecordChange(this.$item.$itemsPerPage = metadata.$itemsPerPage, "$itemsPerPage");
            }
        }
        if (this.$prototype.$cube) {
            if (!this.graphDecorator) {
                this.graphDecorator = new GraphDecorator();
            }
            this.graphDecorator.applyGraphSettings(this, metadata);
        }
        return RawList.prototype.applyDesignMetaData.call(this, metadata, onAuthoring);
    },
    getAuthoringWidget: function(){
        return new authoringOpener.WidgetAuthoring();
    },
    ensureDiagnoseSlot: function(){
        if (!this.$$diagnose && this.$$dataValue) {
            var div = document.createElement("div");
            div.className = "s-field-diagnose";
            this.$$diagnose = $(this.$$dataValue[0].appendChild(div));
        }
        else {
            this.$$diagnose.empty();
        }
    },
    _appendCore: function(){
        var list = this.$$fieldValue[0];
        list.className += " s-list";
        
        this._topbar = document.createElement("div");
        this._topbar.className = "s-list-topbar";
        list.appendChild(this._topbar);
        
        this._core = document.createElement("div");
        this._core.className = "s-list-core";
        list.appendChild(this._core);
        this._appendAuthoringOpener();
        
        this.appendArticleMenus();
        this.appendPager();
    },
    drawBox: function(){
        RawList.prototype.drawBox.call(this);
        var self = this;
        self.$$item.delegate("a[data-s-capability]", "click", function(){
            if (!self.$isDisabled) {
                self.builder.doCapabilities(self.findRecord($(this)), this.getAttribute("data-s-capability"));
            }
            return false;
        }).delegate("[data-s-record]", "mouseenter mouseleave", function(event){
            if (!self.$isDisabled) {
                var isEnter = event.type == "mouseenter";
                var record = self.findRecord($(this));
                record.setCapabilitiesVisible(isEnter);
            }
            return false;
        });
    },
    setState: function(state){
        RawList.prototype.setState.call(this, state);
        if (this._records && this.$capability && state.$isDisabled !== undefined) {
            this._records.forEach(function(record, index){
                record.setState(state);
            });
        }
    },
    applyCapabilities: function(metaData){
        RawList.prototype.applyCapabilities.call(this, metaData);
        if (this.$capability && this.$capability.reorder) {
            if (!this.recordReorder) {
                (this.recordReorder = new RecordReorder()).setDraggable(this);
            }
        }
        else {
            if (this.recordReorder) {
                document.controller.disposeObject(this.recordReorder);
                delete this.recordReorder;
            }
        }
    },
    
    appendArticleMenus: function(){
        var div = document.createElement("div");
        div.className = "s-list-menus-cell";
        this._menus = document.itemFactory.load($(this._topbar.appendChild(div)), {
            $id: this.id + "menus",
            $category: "links",
            $skin: this.$layoutOptions.menus
        }, this);
    },
    appendPager: function(){
        if (!this.$item.$isPagerHidden) {
            (this._pager = new Pager(this)).create(this, this._topbar);
        }
    },
    _appendAuthoringOpener: function(){
        if (this.$item.$isQuickDesignerEnabled !== false) {
            this.quickDesigner = new authoringOpener.QuickDesigner();
            this.quickDesigner.appendOpener(this)
        }
    },
    dispose: function(){
        delete this._menus;
        RawList.prototype.dispose.call(this);
    }
});
