// syracuse-tablet/html/js/ui/dragManager.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var notifications = require('syracuse-tablet/html/js/helpers/notifications');
var defaultOptions = {
	drag: {
		options: {
			containment: ".s-m-main-content",
			disabled: false,
			helper: 'clone',
			opacity: 1,
		}
	},
	drop: {
		options: {
			tolerance: "pointer",
			hoverClass: "s-m-hover-droptarget",
		}
	}
}

var _Klass = utils.defineClass(
	/*
	 * params :
	 * 	dragOpts:{
	 * 		selector: Liste of JQUERY selectors : all elements dragable ex ".s-m-tile,.another selector, ..."	
	 * 		notify:{ all notif usable on drag : start, drag drop avoid drag because to much notif
	 *			start:"sm.tile.start" format : "event": "notif"
			}
			options:{
				jquery drag options
			}
	 * }
	 * dropOpt :{
	 * 	selector: Liste of JQUERY selectors : all elements target ex ".s-m-tile,.another selector, ..."	
	 * 		notify:{ all notif usable on drop : drop, out
	 *			drop:"sm.tile.drop" format : "event": "notif"
			}
			options:{
				jquery drop options
			}
	 * }
	 */
	function DragManager(dragOpts, dropOpt, opts) {
		this.options = opts || {};
		this._notif = _notifier();
		this._init(dragOpts, dropOpt)
	}, null, {
		destroy: function() {
			this._notif = null;
		},
		_init: function(dragOpts, dropOpts) {
			var self = this;
			self.dropOpts = $.extend(true, defaultOptions.drop, dropOpts),
				self.dragOpts = $.extend(true, defaultOptions.drag, dragOpts),
				self._notif.register(self.dragOpts.notify);
			self._notif.register(self.dropOpts.notify);

			$(self.dropOpts.selector).droppable(
				$.extend(true, self.dropOpts.options, {
					drop: function(event, ui) {
						$(this).parent().toggleClass("s-m-not-droppable", true);
						self._notif.publish("drop", $(this));
					},
					out: function(event, ui) {
						$(this).parent().toggleClass("s-m-not-droppable", false);
						self._notif.publish("out", $(this));
					}
				})
			);
			$(self.dragOpts.selector).draggable(
				$.extend(true, self.dragOpts.options, {
					start: function(event, ui) {
						$(ui.helper.context).css("visibility", "hidden");
						ui.helper.css("z-index", 2000);
						self._notif.publish("start", $(this));
					},
					drag: function(event, ui) {

					},
					stop: function(event, ui) {
						$(ui.helper.context).css("visibility", "visible");
						if (self.dragOpts.stopNotify) {
							self._notif.publish("stop", $(this));
						}
					}
				})
			);
		},
		getNotifs: function() {
			return this._notif.get();
		},
	}
);

function _notifier() {
	var notifs = {
		_notifs: {},
		register: function(notifs) {
			var self = this;
			if (!notifs) return;
			Object.keys(notifs).forEach(function(key) {
				self._notifs[key] = notifs[key];
			});
		},
		get: function(event) {
			var self = this;
			if (event) {
				return self._notifs[event]
			} else {
				var notif = [];
				Object.keys(self._notifs).forEach(function(key) {
					notif.push(self._notifs[key]);
				});
				return notif;
			}
		},
		publish: function(event, $$elt) {
			var self = this;
			if (!event || !self.get(event)) return;
			notifications.publish(self.get(event), $$elt);
		}
	};
	return notifs;
}
exports.DragManager = _Klass;
