// syracuse-tablet/html/js/ui/fontUtils.js
"use strict";

var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("app");
var iconList = require('syracuse-tablet/html/js/ui/iconList');

var _paginBtns = {
	"$first": "glyphicon glyphicon-step-backward",
	"$previous": "glyphicon glyphicon-chevron-left",
	"$next": "glyphicon glyphicon-chevron-right",
	"$last": "glyphicon glyphicon-step-forward",
	"$more": "glyphicon glyphicon-refresh"
};

var _getIconByIdx = function(idx) {
	if (idx == null) {
		idx = Math.random() * (iconList.fontAwesomeIconNames.length - 1);
	} else if (idx < iconList.fontAwesomeIconNames.length) {
		idx = Math.abs(idx);
	} else {
		var x = Math.round(Math.abs(idx) / iconList.fontAwesomeIconNames.length);
		idx = idx - x * iconList.fontAwesomeIconNames.length;
	}
	return "fa " + iconList.fontAwesomeIconNames[idx];

};

var _getIconByName = function(name) {
	switch (name) {
		case "$cancelEdit":
			return "fa fa-undo";
		case "$edit":
			return "fa fa-pencil-square-o";
		case "$query":
			return "fa fa-list";
		case "$save":
			return "fa fa-floppy-o";
		case "$details":
			return "fa fa-level-down";
		case "$delete":
			return "fa fa-trash-o";
		case "$create":
			return "fa fa-plus-square";
			break;
		default:
			if (iconList.fontAwesomeIconNames.indexOf("fa-" + name) > -1) {
				return "fa fa-" + name;
			}
			log && log("Icon name=" + name + " not implemented yet");
			return "fa fa-warning";
			break;
	}
	return true;
};

exports.getIconByIdx = _getIconByIdx;
exports.getIconByName = _getIconByName;
exports.checkBoxIcon = function(val) {
	return ["glyphicon", val ? "glyphicon-ok" : "glyphicon-remove"];
};
exports.dateFieldIcon = function($type) {
	return $type.indexOf("date") != -1 ? ["glyphicon", "glyphicon-calendar"] : ["glyphicon", "glyphicon-time"];
};
exports.refFieldIcon = function() {
	return ["glyphicon", "glyphicon-eye-open"];
};
// s-m-img/s-m-binary - icon: default
exports.binaryIcon = function(type, icon) {
	if (type === "s-m-img") {
		return "glyphicon " + (icon === "default" ? "glyphicon-picture" : "");
	}
	if (type === "s-m-binary") {
		return "glyphicon " + (icon === "default" ? "glyphicon-file" : "");
	}
	return "";
};
exports.paginationIcon = function(linkName) {
	return _paginBtns[linkName];
};
exports.pageIcon = function(name) {
	if (name === "clearCache") return "fa fa-history";
	if (name === "switchContext") return "fa fa-user";
	if (name === "search") return "fa fa-search";
	if (name === "sidePanelL") return "fa fa-caret-square-o-left";
	if (name === "sidePanelR") return "fa fa-angle-double-right";
	//if (name === "sidePanelR") return "fa fa-caret-square-o-right";
	if (name === "home") return "fa fa-home";
	if (name === "designPage") return "fa fa-paint-brush";
	if (name === "back") return "fa fa-arrow-circle-o-left";
	if (name === "developmentPage") return "fa fa-gear";
	if (name === "bars") return "fa fa-bars";
	if (name === "backIphone") return "fa fa-angle-left";
	if (name === "about") return "fa fa-info-circle";
	if (name === "logout") return "fa fa-sign-out";
	if (name === "pinPage_1") return "fa fa-star-o";
	if (name === "pinPage_2") return "fa fa-star";
	if (name === "pageConfig") return "fa fa-wrench";
};
exports.ctrlIcon = function(name) {
	if (name === "editrow") return "fa fa-pencil-square-o";
	if (name === "detailrow") return "fa fa-hand-o-right";
	if (name === "delrow") return "fa fa-remove";
	if (name === "addrow") return "fa fa-plus";
};
exports.vignetteIcons = function(name) {
	if (name === "home") return "fa fa-home";
	if (name === "back") return "fa fa fa-arrow-circle-o-left";
	if (name === "refresh") return "fa fa-history";
};

exports.sortIcon = function(name) {
	if (name === "asc") return "fa fa-long-arrow-up";
	if (name === "none") return "fa fa-times";
	if (name === "desc") return "fa fa-long-arrow-down";
};

exports.filterIcon = function(name) {
	if (name === "panel") return "fa fa-bars";
	if (name === "filtered") return "fa fa-filter";
	if (name === "edit") return "fa fa-pencil";
	if (name === "delete") return "fa fa-times";
};
