// syracuse-tablet/html/js/ui/fontUtils.js
"use strict";
var _paginBtns = {
	"$first": "glyphicon glyphicon-step-backward",
	"$previous": "glyphicon glyphicon-chevron-left",
	"$next": "glyphicon glyphicon-chevron-right",
	"$last": "glyphicon glyphicon-step-forward"
};

var _tileIcons = [
	'adjust',
	'adn',
	'ambulance',
	'anchor',
	'android',
	'angellist',
	'apple',
	'archive',
	'asterisk',
	'automobile',
	'ban',
	'bank',
	'barcode',
	'bell',
	'bicycle',
	'binoculars',
	'bitbucket',
	'bolt',
	'book',
	'bookmark',
	'briefcase',
	'building',
	'bullhorn',
	'bullseye',
	'calculator',
	'calendar',
	'camera',
	'certificate',
	'chain',
	'child',
	'circle',
	'clipboard',
	'codepen',
	'cog',
	'cogs',
	'columns',
	'compass',
	'dashboard',
	'database',
	'envelope',
	'exclamation',
	'eye',
	'fax',
	'file',
	'flag',
	'flash',
	'folder',
	'globe',
	'heart',
	'home',
	'image',
	'inbox',
	'info',
	'institution',
	'key',
	'laptop',
	'leaf',
	'legal',
	'lock',
	'magic',
	'money',
	'paperclip',
	'phone',
	'plane',
	'rocket',
	'search',
	'shield',
	'sitemap',
	'sliders',
	'star',
	'steam',
	'stethoscope',
	'stop',
	'suitcase',
	'th',
	'tumblr',
	'twitch',
	'twitter',
	'umbrella',
	'warning',
	'wrench'
];

var _getIconByName = function(name) {
	switch (name) {
		case "$multiSelection":
			return "glyphicon glyphicon-list-alt";
		case "$cancelMultiSel":
		case "$cancelEdit":
			return "fa fa-undo";
		case "$edit":
			return "fa fa-edit";
		case "$query":
			return "fa fa-list";
		case "$save":
			return "fa fa-save";
		case "$details":
			return "fa fa-level-down";
		case "$delete":
			return "fa fa-trash-o";
		case "$create":
			return "fa fa-plus-square";
		case "$image-sign":
			return "fa fa-pencil";
		case "$image-camera":
			return "fa fa-camera";
		case "$image-ink":
			return "fa fa-paint-brush";
		case "appUpdate":
			return "fa fa-history";
		case "appSetup":
			return "fa fa-download";
		case "gear":
			return "fa fa-gear";
		case "expand":
			return "fa fa-expand";
		case "compress":
			return "fa fa-compress";
		case "GPSPicker":
			return "fa fa-dot-circle-o";
		case "scanBarcodePicker":
			return "fa fa-barcode";
		case "penTextInputPicker":
			return "fa fa-pencil-square-o";
		case "$apply":
			return "fa fa-check";
		case "$refresh":
			return "fa fa-save";
		case "lookupPicker":
			return "glyphicon glyphicon-eye-open";
		case "datePicker":
			return "glyphicon glyphicon-calendar";
		case "timePicker":
			return "glyphicon glyphicon-time";
		case "clientTestApplication":
			return "fa fa-th"
	}
	// Icon names used for tiles or tile headers
	if (_tileIcons.indexOf(name) > -1) {
		return "fa fa-" + name;
	}

	return "";
};

exports.getIconByName = _getIconByName;
exports.checkBoxIcon = function(val) {
	return ["glyphicon", val ? "glyphicon-ok" : "glyphicon-remove"];
};
exports.phoneFieldIcon = function() {
	return "fa fa-phone";
};
exports.emailFieldIcon = function() {
	return "fa fa-envelope-o";
};
exports.documentFieldIcon = function() {
	return "fa fa-file-text";
};
exports.binaryFieldIcon = function() {
	return "fa fa-file-0";
};
exports.paginationIcon = function(linkName) {
	return _paginBtns[linkName];
};
exports.layoutExpandCollapseIcon = function(collapse, advancedTheme) {
	if (collapse) {
		return advancedTheme ? "fa fa-ellipsis-h" : "fa fa-chevron-down";
	} else {
		return advancedTheme ? "fa fa-times" : "fa fa-chevron-up";
	}
};

exports.dashboardIcon = function(name) {
	if (name === "removeTile") return "fa fa-trash-o";
	if (name === "editTile") return "fa fa-pencil";
	if (name === "moveTile") return "fa fa-sitemap";
	if (name === "dashboardList") return "fa fa-list-ul";
	if (name === "templateList") return "fa fa-list-ul";
	if (name === "addGroup") return "fa fa-sitemap";
	if (name === "addHiddenTile") return "fa fa-th-large";
	if (name === "moveGroupAfter") return "fa fa-arrow-left";
	if (name === "moveGroupBefore") return "fa fa-arrow-right";
	if (name === "removeGroup") return "fa fa-trash-o";
	if (name === "emptyTile") return "fa fa-exclamation-triangle";
}
exports.tableActionIcon = function(name) {
	if (name === "rowEdit") return "fa fa-pencil-square-o";
	if (name === "rowDetail") return "fa fa-hand-o-right";
	if (name === "rowDelete") return "fa fa-remove";
	if (name === "rowAdd") return "fa fa-plus";
	return "fa fa-question";
};
exports.sortIcon = function(name) {
	if (name === "asc") return "fa fa-long-arrow-up";
	if (name === "none") return "fa fa-times";
	if (name === "desc") return "fa fa-long-arrow-down";
};

exports.filterIcon = function(name) {
	if (name === "panel") return "fa fa-bars";
	if (name === "filtered") return "fa fa-filter";
	if (name === "edit") return "fa fa-pencil";
	if (name === "delete") return "fa fa-times";
	if (name === "separator") return "fa fa-plus";
};
exports.sortFilterTab = function(name) {
	if (name === "sort") return "fa fa-arrows-v";
	if (name === "filter") return "fa fa-filter";
	if (name === "memo") return "fa fa-star";
	if (name === "apply") return "fa fa-check";
};
/* 
 * List of icon names that can be selected by tile customisation
 * These are only the FA icons that look appropriate
 */
exports.getTileIconList = function() {
	return _tileIcons;
};
var _arrayStatusIconNames = ["thumbs-down", "thumbs-o-down", "thumbs-up", "thumbs-o-up", "hand-o-down", "hand-o-up", "lock", "unlock", "unlock-alt", "smile-o", "meh-o", "check", "check-circle", "check-circle-o", "check-square", "check-square-o", "square", "square-o", "circle", "circle-thin", "circle-o", "close", "comment", "comment-o", "comments", "comments-o", "photo", "film", "barcode", "bars", "plus", "plus-circle", "plus-square", "plus-square-o", "minus", "minus-circle", "minus-square", "minus-square-o", "question", "question-circle", "trash", "trash-o", "info", "info-circle", "flag", "flag-checkered", "flag-o", "female", "male", "mars", "venus", "user", "users", "user-plus", "user-times", "user-secret", "facebook-official", "file", "file-o", "dollar", "eur", "gbp", "yen", "rub", "rupee", "shekel", "cny", "ils", "inr", "krw", "money", "turkish-lira", "won", "envelope-square", "envelope-o", "envelope", "exclamation", "exclamation-circle", "exclamation-triangle"];
var _arrayStatusIcons = null;
var _getArrayStatusIconList = function() {
	if (_arrayStatusIcons == null) {
		_arrayStatusIcons = [];
		_arrayStatusIconNames.forEach(function(name) {
			_arrayStatusIcons.push({
				css: "fa fa-" + name,
				name: name
			});
		});
	}
	return _arrayStatusIcons;
};
exports.getArrayStatusIconList = _getArrayStatusIconList;

exports.getArrayStatusDefIcon = function() {
	return {
		name: "circle",
		css: "fa fa-circle"
	};
};

/* Actions control - authoring */
var _arrayActionsIconNames = _arrayStatusIconNames.slice(0);
var _arrayActionsIcons = null;
var _getPageActionIcon = function() {
	if (_arrayActionsIcons == null) {
		_arrayActionsIconNames = _arrayActionsIconNames.concat(["download", "save", "undo", "edit", "history", "eraser", "reply", "refresh"]);
		_arrayActionsIcons = [];
		_arrayActionsIconNames.forEach(function(name) {
			_arrayActionsIcons.push({
				css: "fa fa-" + name,
				name: name
			});
		});
	}
	return _arrayActionsIcons;
};
exports.getPageActionIcon = _getPageActionIcon;

var _emptyImageIconNames = ["image", "user", "users", "warning", "file-movie-o", "file-image-o", "paperclip", "film"];
var _emptyImageIcons = null;
exports.getEmptyImageIconList = function() {
	if (_emptyImageIcons == null) {
		_emptyImageIcons = [];
		_emptyImageIconNames.forEach(function(name) {
			_emptyImageIcons.push({
				css: "fa fa-" + name,
				name: name
			});
		});
	}
	return _emptyImageIcons;
};
exports.getEmptyImageDefIcon = function() {
	return {
		name: "image",
		css: "fa fa-image"
	};
};
exports.getEmptyFileDefIcon = function() {
	return {
		name: "file",
		css: "fa fa-file"
	};
};

var _urlIconNames = ["globe", "map-marker", "street-view", "road", "link", "cloud", "database", "bed", "cutlery", "glass", "train", "subway", "taxi", "ship", "plane"];
var _urlIcons = null;
exports.getUrlIconList = function() {
	if (_urlIcons == null) {
		_urlIcons = [];
		_urlIconNames.forEach(function(name) {
			_urlIcons.push({
				css: "fa fa-" + name,
				name: name
			});
		});
	}
	return _urlIcons;
};
exports.getUrlDefIcon = function() {
	return {
		name: "Url",
		css: "fa fa-globe"
	};
};
var _colImageIconNames = ["image", "user", "users", "warning", "file-movie-o", "file-image-o", "paperclip", "film"];
var _colImageIcons = null;
exports.getColImageIconList = function() {
	if (_colImageIcons == null) {
		_colImageIcons = [];
		_colImageIconNames.forEach(function(name) {
			_colImageIcons.push({
				css: "fa fa-" + name,
				name: name
			});
		});
	}
	return _colImageIcons;
};
exports.getColImageDefIcon = function(emptyImage) {
	return {
		name: "image",
		css: emptyImage === true ? "fa fa-times" : "fa fa-image"
	};
};
exports.getCarouselIcon = function(what) {
	return what == "left" ? "glyphicon glyphicon-chevron-left" : what == "right" ? "glyphicon glyphicon-chevron-right" : "";
};


var _panelActionIconMap = {
	"home": "fa fa-home",
	"pin": "fa fa-star-o",
	"unpin": "fa fa-star",
	"openactionspanel": "fa fa-ellipsis-h",
	"switchContext": "fa fa-caret-down",
	"mail": "fa fa-envelope",
	"share": "fa fa-share-alt",
	"settings": "fa fa-gear",
	"about": "fa fa-info-circle",
	"development": "fa fa-wrench",
	"logout": "fa fa-sign-out",
	"drafts": "fa fa-building-o",
	"designPage": "fa fa-thumb-tack",
	"nativeShare": "fa fa-share-alt",
	"accept": "fa fa-check",
	"cancel": "fa fa-reply",

	"$save": "fa fa-save",
	"$edit": "fa fa-edit",
	"$delete": "fa fa-trash-o",
	"$create": "fa fa-plus-square",
	"$query": "fa fa-list",

	"multiSelection": "fa fa-arrow-left",
	"multiSelectionRunExit": "fa fa-times",
	"multiSelectionRunStop": "fa fa-stop",

	"search": "fa fa-search",
	"sortFilter": "fa fa-filter",

	"pageConfig": "fa fa-pencil",
	"addGroup": "fa fa-sitemap",
	"myDashboardTemplates": "fa fa-list-ul",
	"myAvailableTemplates": "fa fa-list-ul",
	"myDashboard": "fa fa-chevron-left",
	"showEditTemplates": "fa fa-file-text-o",
	"showEditTemplatesEmpty": "fa fa-file-o",
	"draftsRunStop": "fa fa-stop",
	"draftsRunSynch": "fa fa-upload",
	"draftsRemove": "fa fa-trash"
};

exports.panelActionIcon = function(name) {
	return _panelActionIconMap[name] || "";
}
