// syracuse-tablet/html/js/ui/modalSaveAuth.js
"use strict";

var modal = require('syracuse-tablet/html/js/ui/modal');
var locale = require('syracuse-tablet/html/js/helpers/locale');

var _template = '\
	<div class="modal fade" id="modal-endpoint" tabindex="-1" role="dialog"> \
    <div class="modal-dialog"> \
		<div class="modal-content"> \
		    <div class="modal-header"> \
				<h4 class="modal-title">{{title}}</h4> \
				<label>{{page}}</label> \
		    </div> \
		    <div class="modal-body"> \
				<div class="form-group"> \
					<label>{{label_code}}</label> \
					<input class="form-control" id="authCode" value="{{authCodeDefault}}" /> \
				</div> \
				<div class="form-group"> \
					<label>{{label_title}}</label> \
					<input class="form-control" id="authTitle" value="{{authTitleDefault}}" /> \
				</div> \
				<div class="form-group"> \
					<label>{{label_description}}</label> \
					<input class="form-control" id="authDescription" value="{{authDescriptionDefault}}"  /> \
				</div> \
				<label>{{label_saveas_option}}</label> \
				<div class="btn-group btn-group-justified" data-toggle="buttons"> \
					<div class="btn-group"> \
						<label class="btn btn-primary"> \
							<input type="radio" name="options" id="personal_copy" checked>{{label_personal}} \
						</label> \
					</div> \
					<div class="btn-group"> \
						<label class="btn btn-primary"> \
							<input type="radio" name="options" id="global_variant"> {{label_global}} \
						</label> \
					</div> \
					<div class="btn-group"> \
						<label class="btn btn-primary active"> \
							<input type="radio" name="options" id="factory_variant">{{label_factory}} \
						</label> \
					</div> \
				</div> \
			</div> \
		    <div class="modal-footer"> \
		        <button type="button" class="btn" data-action="validate">{{label_save}}</button> \
    			<button type="button" class="btn btn-default" data-action="cancel">{{label_cancel}}</button> \
			</div> \
		</div> \
	</div> \
	</div>';

/**
 */
var _selectContext = function(context) {
	var deferred = $.Deferred();
	try {
		try {
			var modalHtml = Handlebars.compile(_template);
			var ctx = {
				title: locale.text("auth.save.title"),
				page: context.pageName,
				authCodeDefault: context.code,
				authTitleDefault: context.title,
				authDescriptionDefault: context.description,

				label_code: locale.text("auth.panel.label_code"),
				label_title: locale.text("auth.panel.label_title"),
				label_description: locale.text("auth.panel.label_description"),
				label_saveas_option: locale.text("auth.panel.label_saveas_option"),
				label_personal: locale.text("auth.panel.label_personal"),
				label_global: locale.text("auth.panel.label_global"),
				label_factory: locale.text("auth.panel.label_factory"),

				label_save: locale.text("auth.panel.label_save"),
				label_cancel: locale.text("auth.panel.label_cancel"),
			};
			if (context) {}

			modalHtml = modalHtml(ctx);
			modal.modal(modalHtml, function(msg, dialg) {
				var code = dialg.find("#authCode").val();
				var title = dialg.find("#authTitle").val();
				var description = dialg.find("#authDescription").val();
				if (msg === "validate") {
					deferred.resolve({
						action: "ok",
						code: code,
						title: title,
						description: description,
						saveAs: $("input:radio:checked", dialg)[0].id
					});
				} else if (msg === "cancel") {
					deferred.resolve({
						action: "cancel"
					});
				}
			});
		} catch (e) {
			deferred.reject(e);
		}
	} catch (e) {
		deferred.reject(e);
	} finally {
		return deferred.promise();
	}
};

exports.selectContext = _selectContext;
