// syracuse-tablet/html/js/ui/modalChooseApps.js
"use strict";

var modal = require('syracuse-tablet/html/js/ui/modal');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var globals = require('syracuse-tablet/html/js/helpers/globals');

var _template = '\
	<div class="modal fade" id="modal-endpoint" tabindex="-1" role="dialog"> \
    <div class="modal-dialog"> \
		<div class="modal-content"> \
		    <div class="modal-header"> \
				<h4 class="modal-title">{{title}}</h4> \
		    </div> \
		    <div class="modal-body"> \
				<div class="form-group"> \
					<div class="list-group" id="modal-device-select-list-id"> \
						{{#each devices}} \
							<a href="#" class="list-group-item" data-auth-device="{{name}}">{{label}}</a> \
						{{/each}} \
					</div> \
				</div> \
		    </div> \
		    <div class="modal-footer"> \
				<button type="button" class="btn btn-default" data-action="ok">{{label_add}}</button> \
    			<button type="button" class="btn btn-default" data-action="cancel">{{label_cancel}}</button> \
			</div> \
		</div> \
	</div> \
	</div>';

/**
 */
var _chooseApps = function(appList) {
	var deferred = $.Deferred();
	try {
		var modalHtml = Handlebars.compile(_template);
		var ctx = {
			title: locale.text("welcome.select.app.title"),
			devices: appList,
			label_add: locale.text("welcome.select.app.add"),
			label_cancel: locale.text("welcome.select.app.cancel"),
		};

		modalHtml = modalHtml(ctx);
		modal.modal(modalHtml, function(msg, $$dialg) {

			$$dialg.find("#modal-device-select-list-id a").on("click", function() {
				$$dialg.modal("hide");
				deferred.resolve({
					action: "ok",
					apps: []
				});
			});
			if (msg === "cancel") {
				deferred.resolve({
					action: "cancel"
				});
			}
		});
	} catch (e) {
		deferred.reject(e);
	} finally {
		return deferred.promise();
	}
};

exports.chooseApps = _chooseApps;
