// syracuse-tablet/html/js/ui/modalChooseApps.js
"use strict";

var globals = require('syracuse-tablet/html/js/helpers/globals');
var utils = require('syracuse-tablet/html/js/helpers/utils');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var modalDialog = require('syracuse-tablet/html/js/ui/modalDialog');

var _templateTablet = '\
	<div class="modal fade" id="modal-endpoint" tabindex="-1" role="dialog"> \
    <div class="modal-dialog"> \
		<div class="modal-content"> \
		    <div class="modal-header"> \
				<h4 class="modal-title">{{title}}</h4> \
		    </div> \
		    <div class="modal-body"> \
				<div class="form-group"> \
					<div class="list-group">\
						{{#each appList}} \
							<a href="#" class="list-group-item" data-action="clickApp" data-params="{{applicationName}}">\
								<h4 class="list-group-item-heading">{{title}}</h4>\
								<p class="list-group-item-text">{{description}}</p>\
							</a>\
						{{/each}} \
					</div> \
				</div> \
		    </div> \
		    <div class="modal-footer"> \
				<button type="button" class="btn btn-default" data-action="$validate">{{label_add}}</button> \
    			<button type="button" class="btn btn-default" data-action="$cancel">{{label_cancel}}</button> \
			</div> \
		</div> \
	</div> \
	</div>';

var _templateSmartphone = '\
	<div class="modal fade" id="modal-endpoint" tabindex="-1" role="dialog"> \
    <div class="modal-dialog"> \
		<div class="modal-content"> \
		    <div class="modal-header"> \
				<h4 class="modal-title">{{title}}xxx</h4> \
		    </div> \
		    <div class="modal-body"> \
				<div class="form-group"> \
					<div class="list-group">\
						{{#each appList}} \
							<a href="#" class="list-group-item" data-action="clickApp" data-params="{{applicationName}}">\
								<h4 class="list-group-item-heading">{{title}}</h4>\
								<p class="list-group-item-text">{{description}}</p>\
							</a>\
						{{/each}} \
					</div> \
				</div> \
		    </div> \
		    <div class="modal-footer"> \
				<button type="button" class="btn btn-default" data-action="$validate">{{label_validate}}</button> \
    			<button type="button" class="btn btn-default" data-action="$cancel">{{label_cancel}}</button> \
			</div> \
		</div> \
	</div> \
	</div>';

var _Klass = utils.defineClass(
	function ApplicationSelectionDialog(appList) {
		var self = this;
		modalDialog.ModalBase.call(self);
		// Init params
		self.appList = appList;

		// Result
		self.selectedApps = {};
	}, modalDialog.ModalBase, {
		_getTemplateStrings: function() {
			var self = this;
			return {
				std: _templateTablet,
				tablet: _templateTablet,
				smartphone: _templateSmartphone
			};
		},

		_getDataContext: function() {
			var self = this;
			return {
				label_validate: "OK",
				label_cancel: "Cancel",
				title: "Test modal",
				appList: self.appList
			};
		},

		// could be removed, result is set in _onAction
		_onValidate: function() {},
		_onClose: function() {},

		_onAction: function(action, param) {
			var self = this;
			if (action === "clickApp") {
				var $$app = self.$$elmt.find("[data-params=" + param + "]");
				if (self.selectedApps[param]) {
					$$app.removeClass("active");
					delete self.selectedApps[param];
				} else {
					$$app.addClass("active");
					self.selectedApps[param] = true;
				}
			}
			// dont close on actions
			return false;
		},
		_getResult: function() {
			var self = this;
			return Object.keys(self.selectedApps);
		}
	}
);

exports.Modal = _Klass;
