// syracuse-tablet/html/js/ui/modalConfigTile.js
"use strict";

var globals = require('syracuse-tablet/html/js/helpers/globals');
var utils = require('syracuse-tablet/html/js/helpers/utils');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var modalDialog = require('syracuse-tablet/html/js/ui/modalDialog');
var iconList = require('syracuse-tablet/html/js/ui/iconList');
var HammerScroller = require('syracuse-tablet/html/js/helpers/hammerScroller').Klass;
var uiRect = require('syracuse-tablet/html/js/ui/rect');

var authoringProperties = require('syracuse-tablet/html/js/authoring/authoringProperties');

var _templateTablet = {
	header: '\
			<h4 class="modal-title">{{title}}</h4> \
			',
	content: ' \
			<div class="s-m-scroll-wrapper s-m-scroll-wrapper-v">\
			<div class="form-group s-m-scroll-element"> \
				<label>{{lab_title}}</label> \
				<input id="tile_title"> \
				<label>{{lab_color}}</label> \
				<div class="s-m-scroll-wrapper s-m-scroll-wrapper-h">\
				<div class="s-m-list-colors s-m-scroll-element" style="width: {{colorsWidth}}px;">\
					{{#each colors}} \
						<a href="#" class="s-m-bg-{{value}}" data-action="clickColor" data-params="{{value}}">\
						</a>\
					{{/each}} \
				</div> \
				</div> \
				<label>{{lab_size}}</label> \
				<div class="s-m-scroll-wrapper s-m-scroll-wrapper-h">\
				<div class="s-m-list-sizes s-m-scroll-element" style="width: {{sizesWidth}}px;">\
					{{#each sizes}} \
						<a href="#" class="" data-action="clickSize" data-params="{{value}}">{{label}}</a>\
					{{/each}} \
				</div> \
				</div> \
				<label>{{lab_icon}}</label> \
				<div class="s-m-scroll-wrapper s-m-scroll-wrapper-h">\
				<div class="s-m-list-icons s-m-scroll-element" style="width: {{iconsWidth}}px;">\
					{{#each icons}} \
						<a href="#" data-action="clickSize" data-params="{{value}}"> \
							<i class="fa {{value}}"/> \
						</a>\
					{{/each}} \
				</div> \
				</div> \
			</div> \
			</div>',

	footer: '\
			<button type="button" class="btn btn-default" data-action="$validate">{{label_validate}}</button> \
			<button type="button" class="btn btn-default" data-action="$cancel">{{label_cancel}}</button> \
			',
};

var _Klass = utils.defineClass(
	function ApplicationSelectionDialog(tile) {
		var self = this;
		modalDialog.ModalBase.call(self);

		// Init params
		self.tile = tile;
		// Result
		self.result = null;
	}, modalDialog.ModalBase, {
		_getTemplates: function() {
			var self = this;
			return {
				std: _templateTablet,
				// tablet: "alternative 1",
				// smartphone:"alternative 2",
			};
		},
		getDisplayFlags: function() {
			var self = this;
			var displayFlags = self.displayFlags || {};
			if (self.deviceType === "tablet") {
				displayFlags.modalClass = "config-tile half_center";
			} else {
				displayFlags.modalClass = "config-tile full";
			}
			return displayFlags;
		},

		_getDataContext: function() {
			var self = this;
			authoringProperties.initProperties();

			var properties = authoringProperties.getAllProperties();
			var data = {
				label_validate: locale.text("welcome.tile.cfg.ok"),
				label_cancel: locale.text("welcome.tile.cfg.cancel"),
				title: locale.text("welcome.tile.cfg.title"),

				lab_title: locale.text("welcome.tile.cfg.lab_title"),
				lab_color: locale.text("welcome.tile.cfg.lab_color"),
				lab_size: locale.text("welcome.tile.cfg.lab_size"),
				lab_icon: locale.text("welcome.tile.cfg.lab_icon")
			};

			data.colors = properties.tileColor.values.map(function(e) {
				return e;
			});
			data.colors.shift(); // First one is transparent which we do not want here

			data.sizes = properties.tileSize.values.map(function(e) {
				return e;
			});

			data.icons = iconList.fontAwesomeIconNames.map(function(e) {
				return {
					value: e,
					label: e
				};
			});

			data.colorsWidth = data.colors.length * 54;
			data.sizesWidth = data.sizes.length * 94;
			data.iconsWidth = ((data.icons.length + 2) / 3) * 50;

			return data;
		},

		_onShow: function() {
			var self = this;
			modalDialog.ModalBase.prototype._onShow.call(self);

			var $$scrolls = self.$$elmt.find(".s-m-scroll-element");
			var i;
			var scrollerId = 1;
			var scrollers = {};

			for (i = 0; i < $$scrolls.length; i++) {
				scrollerId++;

				var $$se = $($$scrolls[i]);
				var $$wrapper = $$se.parent();
				var dir = $$wrapper.hasClass("s-m-scroll-wrapper-h") ? "h" : "v";

				$$wrapper.attr("data-scroller-id", scrollerId);

				var parentScroller = null;
				var $$parentScroller = $$wrapper.parent().closest(".s-m-scroll-wrapper");
				if ($$parentScroller.length > 0) {
					parentScroller = scrollers[$$parentScroller.attr("data-scroller-id")];
				}

				var css = {};
				css["overflow-" + dir] = "auto";
				$$wrapper.css(css);

				var gestureMgr = new HammerScroller($$se, {
					direction: dir,
					valMax: 0,
					name: "MODAL" + scrollerId,
					isPageScroller: false
				}, parentScroller);

				scrollers[scrollerId] = gestureMgr;

				var scrollRect = uiRect.elmtRect($$se, "outer");
				var wrapperRect = uiRect.elmtRect($$wrapper, "outer");
				var viewRect = wrapperRect.intersectRect(scrollRect);
				if (viewRect && !viewRect.contains(scrollRect)) {
					gestureMgr.init(viewRect);
				} else {
					gestureMgr.reset();
				}
			}
		},

		// could be removed, result is set in _onAction
		_onValidate: function() {

		},
		_onCancel: function() {
			var self;
		},

		_onAction: function(action, param) {
			var self = this;
			if (action === "clickApp") {
				var $$app = self.$$elmt.find('[data-params="' + param + '"]');
				if (self.selectedApps[param]) {
					$$app.removeClass("active");
					delete self.selectedApps[param];
				} else {
					$$app.addClass("active");
					self.selectedApps[param] = true;
				}
			}
			// dont close on actions
			return false;
		},

		_getResult: function() {
			var self = this;
			return self.result;
		}
	}
);

exports.Modal = _Klass;
