// syracuse-tablet/html/js/ui/modal.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var _log = require('syracuse-tablet/html/js/helpers/logger').getLogger("storage");
var templates = require('syracuse-tablet/html/js/helpers/templating');

var _error = function(title, e, cb) {
	if (utils.isError(e)) {
		e = helpers.toDiagnose({
			where: title,
			exception: e
		});
	} else if (typeof e === "string") {
		e = utils.toDiagnose({
			where: title,
			message: e
		});
	} else {
		e = utils.toDiagnose(e);
	}
	var body = JSON.stringify(e, null, 2);
	_log && _log("exports.displayError", body);
	_modal("error", {
		close: "Close",
		title: title || "Error",
		body: body
	}, cb);
};

var _info = function(title, text, cb) {
	_modal("info", {
		close: "Close",
		title: title || "Warning",
		body: text
	}, cb);
};

var _confirm = function(action, cb) {
	_modal("confirm", {
		title: "Confirm",
		action: action,
		yes: "yes",
		no: "no"
	}, function(action) {
		if (cb) cb("yes" === action);
	});
};

var _modal = function(id, context, cb) {
	id = "modals_" + id;
	$.when(templates.execSync(id, context)).always(
		function(html) {
			if (html == null) {
				alert("Modal not found - id=" + id + "\n" + JSON.stringify(context, null, 2));
				if (cb) cb(null, "hidden.bs.modal");
			} else {
				var h = $(html).modal({
					keyboard: true,
					backdrop: "static"
				});
				var done = false;
				// Use <button data-dismiss"modal" for close action
				// Use <button data-action"myaction" for action other than
				// close
				h.on("shown.bs.modal", function() {
					if (cb) cb(h, "shown.bs.modal");
				});
				h.on("hidden.bs.modal", function() {
					// Called when dialog is close - data-dismiss = "modal"
					// or m.modal('hide')
					h.unbind();
					h.remove();
					// Call cb only if no click action
					if (!done && cb) cb(h, "hidden.bs.modal");
				});
				// For dialogs with option
				h.delegate("button[data-action]", "click", function(evt) {
					var t = $(evt.target);
					var act = t.attr("data-action");
					if (act) {
						done = true;
						evt.preventDefault();
						evt.stopPropagation();
						if (cb) cb(h, act, t.attr("data-params"));
					}
					// Close modal
					h.modal('hide');
				});
			}
		});
};
exports.init = function(rooturl) {
	if (rooturl) {
		templates.setConfig(null, rooturl);
	}
	templates.preload(["modals/error", "modals/confirm", "modals/info"]);
};
exports.error = _error;
exports.info = _info;
exports.modal = _modal;
exports.confirm = _confirm;
