// syracuse-tablet/html/js/ui/modalChooseDevice.js
"use strict";

var modal = require('syracuse-tablet/html/js/ui/modal');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var globals = require('syracuse-tablet/html/js/helpers/globals');

var _template = '\
	<div class="modal fade" id="modal-endpoint" tabindex="-1" role="dialog"> \
    <div class="modal-dialog"> \
		<div class="modal-content"> \
		    <div class="modal-header"> \
				<h4 class="modal-title" id="myModalLabel">{{title}}</h4> \
		    </div> \
		    <div class="modal-body"> \
				<div class="form-group"> \
					<div class="list-group" id="modal-device-select-list-id"> \
						{{#each devices}} \
							<a href="#" class="list-group-item" data-auth-device="{{name}}">{{label}}</a> \
						{{/each}} \
					</div> \
				</div> \
		    </div> \
		    <div class="modal-footer"> \
    			<button type="button" class="btn btn-default" data-action="cancel">{{label_cancel}}</button> \
			</div> \
		</div> \
	</div> \
	</div>';

/**
 */
var _chooseDevice = function(device) {
	var deferred = $.Deferred();
	try {
		var devices = [];

		var site = globals.getSiteLayout();
		var devs = site.getDeviceTemplates();
		for (var dev in devs) {
			devices.push({
				name: dev,
				label: locale.textOpt("auth.device." + dev + ".title") || dev
			});
		}

		var modalHtml = Handlebars.compile(_template);
		var ctx = {
			title: locale.text("auth.device.title"),
			devices: devices,
			label_cancel: locale.text("auth.device.cancel"),
		};

		modalHtml = modalHtml(ctx);
		modal.modal(modalHtml, function(msg, $$dialg) {

			$$dialg.find("#modal-device-select-list-id a").on("click", function() {
				var device = $(this).attr("data-auth-device");
				$$dialg.modal("hide");
				deferred.resolve({
					action: "ok",
					device: device
				});
			});
			if (msg === "cancel") {
				deferred.resolve({
					action: "cancel"
				});
			}
		});
	} catch (e) {
		deferred.reject(e);
	} finally {
		return deferred.promise();
	}
};

exports.chooseDevice = _chooseDevice;
