// syracuse-tablet/html/js/ui/uiSettings.js
"use strict";
var jsutils = require('syracuse-tablet/html/js/helpers/jsutils');

var _cache = {};
var _lookAndFeel;

/**
 * typeObject 	hubLayout/footer/header/..
 * path			height/groupSpacing...
 * device		tablet/smartphone
 * orientation	portrait/landscape
 */
exports.init = function(os) {
	_cache = {};
	_lookAndFeel = os;

	_settings.gen = $.extend(true, {}, _settings.def, _settings[os]);
};

/**
 * 
 */
exports.getLookAndFeel = function() {
	return _lookAndFeel;
};

/**
 * 
 */
exports.getProp = function(path, device, orientation) {
	if (path == null || arguments.length === 0) return null;
	var res = jsutils.getPropByPath(_settings.gen, path, _cache);
	if (res == null) return null;
	if (device && res[device]) {
		res = res[device];
	}
	if (orientation && res[orientation]) {
		res = res[orientation];
	}
	if (typeof res === "function") {
		res = res.call(null, device, orientation);
		_cache[Array.isArray(path) ? path.join('.') : path] = res;
	}
	return res;
};

/**
 * SETTINGS
 */
var _settings = {
	ios: null,
	android: null,
	windows: null,
	def: {
		hubLayout: {
			groupSpacing: 55,
			landscapeMinHeigth: function() {
				return exports.getProp("hubGroup.landscapeBaseSize") * 4 + exports.getProp("hubGroup.landscapePadY") * 5 + 40;
			}
		},
		hubGroup: {
			// Default Base size
			// Allows 8 cells from left to right on Nexus 7 portrait mode, 
			// all other devices use a bigger, recalculated base sizes
			landscapeBaseSize: 66,
			landscapePadX: 8,
			landscapePadY: 8,
			portraitPadX: 8,
			portraitPadY: 8,
			portraitWideMinWidth: 570
		},
		footer: {
			height: {
				tablet: 50,
				smartphone: 62
			}
		},
		header: {
			height: 50,
		},
		navigationStackBar: {
			height: 40,
		},
		breadcrumbs: {
			height: 25
		},
		chartDetailToolbar: {
			height: 40
		},
		colors: {
			CELLDEFCOLOR: "darkgrey",
			TILEDEFCOLOR: "darkgrey"
		}
	}
};
