// etna/lib/supervisor/protocol/page.ts
import { FilterProto, X3FormatProto, PropertiesProto, LinksProto, ActionProto } from './prototype';

export interface Layout {
	$items?: Article[];
}

export interface Article {
	$layout?: Layout;
	$category?: string;
	$title?: string;
	$bind?: string;
	$order?: number;
	$kind?: string;
	$menus?: Menu;
}

export interface Menu extends Article {
	$CODMENU?: string;
	$PARMENU?: string;
	$isBoxCollapsable?: boolean;
	$opened?: boolean;
}

export interface PageProto {
	$article?: Article;
	$filters?: { [name: string]: FilterProto; };
	$actxUser?: string;
	$actxLogin?: string;
	$actxFolder?: string;
	$actxSolution?: string;
	$actxLan?: string;
	$actxLanDef?: string;
	$actxLanIso?: string;
	$actxLegCur?: string;
	$actxNbLeg?: string;
	$baseType?: string;
	$type?: string;
	$representation?: string;
	$prototype?: string;
	$device?: string;
	$title?: string;
	$collaborationUrl?: string;
	$baseUrl?: string;
	$repUrl?: string;
	$instanceUrl?: string;
	$x3FormatExt?: { [name: string]: X3FormatProto; };
	$url?: string;
	$itemsPerPage?: number;
	$properties?: PropertiesProto;
	$links?: LinksProto;
	$actions?:  { [name: string]: ActionProto; };
}
