// etna/lib/supervisor/protocol/prototype.ts
import { DiagnosisProto } from './data';

export interface FilterProto {
}

export interface ReturnProto {
}

export interface NodeLink {
	$type?: string;
	$title?: string;
	$url?: string;
	$target?: string;
}

export interface LinkNodeProto {
	$target?: string;
	$method?: string;
	$url?: string;
	$title?: string;
	$type?: string;
	$confirm?: string;
	$officeAddinSetup?: string;
	$parameters?: LinkParamsNodeProto;
	$kind?: string;
	$isHidden?: boolean;
	$isDisabled?: boolean;
	$links?: { [name: string]: LinkNodeProto; };
	$debug?: any;
	$invocationMode?: string;
	$capabilities?: string;
}

export interface ParamPropertiesNodeProto {
	$type?: string;
	$title?: string;
	dummy?: any; // $ missing !!
}

export interface LinkParamsNodeProto {
	$method?: string;
	$url?: string;
	$properties?: ParamPropertiesNodeProto;
	msoCurrentRepresentation?: string; // $ missing!!
}

export interface PropertyRefProto {
	$url: string;
	$value: any;
	$key: string;
	$properties: { [name: string]: NodeProto }
}

export interface NodeProto {
	$url?: string;
	$type?: string;
	$key?: string;
	$value?: any;
	$title?: string;
	$shortTitle?: string;
	$description?: string;
	$isReadOnly?: boolean;
	$isMandatory?: boolean;
	$isHidden?: boolean;
	$isExcluded?: boolean;
	$isDisabled?: boolean;
	$capabilities?: string;
	$format?: string;
	$maxLength?: number;
	$precision?: number;
	$scale?: number;
	$minItems?: number;
	$maxItems?: number;
	$item?: NodeProto;
	$properties?: { [name: string]: NodeProto };
	$instanceUrl?: string;
	$x3Format?: string;
	$unit?: string;
	$acceptedTypes?: string[];
	$links?: {
		$localize?: NodeLink;
		[name: string]: NodeLink; 
	};
	[name: string]: any;
}

export interface PropertyProto {
	$item?: any;
	$type?: string;
	$capabilities?: string;
}

export interface PropertiesProto {
	$resources?: PropertyProto;
	[name: string]: PropertyProto;
}

export interface LinkProto {
	$url?: string;
	$type?: string;
	$method?: string;
	$title?: string;
	$target?: string;
	$isHidden?: boolean;
	$isDisabled?: boolean;
}

export interface LinksProto {
	[name: string]: LinkProto; 
}

export interface ActionProto {
	$diagnoses?: DiagnosisProto[];
	$return?: ReturnProto;
	$isDisabled?: boolean;
	$isHidden?: boolean;
	$isRequested?: boolean;
	$links?: LinksProto;
}

export interface ActionsProto {
	$save?: ActionProto;
	[name: string]: ActionProto;
}

export interface ServiceProto {
	$baseUrl?: string;
	$prototype?: string;
	$representation?: string;
	$url?: string;
	$type?: string;
	$etag?: number;
	ETag?: number;
	$uuid?: string;
	$actions?: ActionsProto;
	$diagnoses?: DiagnosisProto[];
}

export interface X3FormatProto {
	stdType: string;
	chars: string;
}
