// syracuse-tablet/html/js/authoring/authoringHtml.js
"use strict";

/*
 * Authoring panel main layout
 */
var _templates = {
	authPanelRightStructure: '\
		<div class="s-m-site-auth-panel-content"> \
			<ul class="nav nav-pills" role="tablist"> \
				<li><a href="#authPanelControl" role="tab" data-toggle="tab">{{label_control}}</a></li> \
				<li class="active"><a href="#authPanelEdit" role="tab" data-toggle="tab">{{label_edit}}</a></li> \
				<li><a href="#authPanelSave" role="tab" data-toggle="tab">{{label_save}}</a></li> \
			</ul> \
			<div class="tab-content"> \
				<div id="authPanelControl" class="tab-pane panel-body active"> \
		 			<div class="s-m-auth-prop-list"> \
						<label class="control-label">{{label_control_title}}</label> \
				    	<p class="form-control-static" id="authPanelControlTitle"> </p> \
						<label class="control-label">{{label_control_type}}</label> \
						<p class="form-control-static" id="authPanelControlType"> </p> \
					</div> \
					<div id="authPanelControlSimple"> \
					</div> \
				</div> \
				<div id="authPanelEdit" class="tab-pane panel-body"> \
					<label id="authEditName"></label> \
					<textarea  id="authJSONPage" type="text"></textarea> \
					<button type="button" id="authPanelBtnPageApply" class="btn pull-right">{{label_apply}}</button> \
				</div> \
				<div id="authPanelSave" class="tab-pane panel-body"> \
					<label id="authPageName"></label> \
					<div class="form-group"> \
						<label>{{label_code}}</label> \
						<input class="form-control" id="authCode" value="{{authCodeDefault}}" /> \
					</div> \
					<div class="form-group"> \
						<label>{{label_title}}</label> \
						<input class="form-control" id="authTitle" value="{{authTitleDefault}}" /> \
					</div> \
					<div class="form-group"> \
						<label>{{label_description}}</label> \
						<input class="form-control" id="authDescription" value="{{authDescriptionDefault}}"  /> \
					</div> \
					<label>{{label_saveas_option}}</label> \
					<div class="btn-group btn-group-justified" data-toggle="buttons"> \
						<div class="btn-group"> \
							<label class="btn btn-primary"> \
								<input type="radio" name="options" id="personal_copy" checked>{{label_personal}} \
							</label> \
						</div> \
						<div class="btn-group"> \
							<label class="btn btn-primary"> \
								<input type="radio" name="options" id="global_variant"> {{label_global}} \
							</label> \
						</div> \
						<div class="btn-group"> \
							<label class="btn btn-primary active"> \
								<input type="radio" name="options" id="factory_variant">{{label_factory}} \
							</label> \
						</div> \
					</div> \
					<button type="button" id="authPanelBtnSave" class="btn pull-right">{{label_save}}</button> \
				</div> \
			</div> \
		</div>',

	authPanelLeftStructure: '\
			<div class="s-m-site-auth-panel-content"> \
			no content yet...\
			</div>',

	authPanelPropertyGroup: ' \
		<div class="panel panel-primary s-m-panel-auth-prop"> \
			<div class="panel-heading" role="tab" id="heading_{{id}}"> \
				<h4 class="panel-title"> \
					<a data-toggle="collapse" data-parent="#accordion" href="#collapse_{{id}}" {{#if expanded}} aria-expanded="true"{{/if}} aria-controls="collapse_{{id}}" {{#unless expanded}}class="collapsed"{{/unless}}>{{label}}</a> \
				</h4> \
			</div> \
			<div id="collapse_{{id}}" class="panel-collapse collapse {{#if expanded}}in{{/if}}" role="tabpanel" aria-labelledby="heading_{{id}}"> \
			</div> \
		</div>',

	authPanelPropertyRadio: ' \
		<div class="btn-group" data-toggle="buttons" style="width: 100%;"> \
			{{#each options}} \
				<div class="btn-group" style="width: 100%; display: {{this.display}};" id="{{this.id}}"> \
					<label class="s-m-auth-btn-radio-icon s-m-auth-btn-block" id="{{this.id}}_lab" data-s-m-auth-value="{{this.value}}"> \
						<input type="radio" name="{{group}}" id="{{this.id}}">{{this.label}} \
					</label> \
				</div> \
			{{/each}} \
		</div>',

	authPanelPropertyRadioColor: ' \
		<div class="btn-group" data-toggle="buttons" style="width: 100%;"> \
			{{#each options}} \
				<div class="btn-group" style="width: 100%;"> \
					<label class="s-m-auth-btn-radio-icon s-m-auth-btn-block" id="{{this.id}}_lab" data-s-m-auth-value="{{this.value}}"> \
						<input type="radio" name="{{group}}" id="{{this.id}}">{{this.label}} \
						<div class="s-m-auth-btn-radio-color s-m-bg-{{color}} "></div> \
					</label> \
				</div> \
			{{/each}} \
		</div>',

	authPanelPropertyInput: ' \
			<div class="form-group"> \
				<input class="form-control" id="{{id}}" value="{{value}}" /> \
			</div>',

	authPanelPropertyCellSize: ' \
		{{#each sizes}} \
			<label class="{{class}}">{{this.label}}</label> \
			<div class="btn-group x_btn-group-justified" data-toggle="buttons" style="width: 100%;"> \
				{{#each ../values}} \
					<div class="btn-group"> \
						<label class="btn btn-default btn-lg" id="{{this.id}}_lab" data-s-m-auth-prop="{{../this.prop}}" data-s-m-auth-value="{{this}}"> \
							<input type="radio" name="{{group}}" id="{{../this.prop}}={{this}}">{{this}} \
						</label> \
					</div> \
				{{/each}} \
			</div> \
		{{/each}}',

	authPanelPropertyJSON: ' \
		<div class="form-group"> \
			<textarea class="form-control" id="{{id}}" rows="10">{{value}}</textarea> \
		</div>'
};

var _compiled = {};

function _getCompiled(name) {
	var c = _compiled[name];
	if (c) {
		return c;
	}

	var t = _templates[name];
	c = Handlebars.compile(t);
	_compiled[name] = c;
	return c;
}

exports.execute = function(name, ctx) {
	var c = _getCompiled(name);
	return c(ctx);
};
