// syracuse-tablet/html/js/authoring/authoringHelpers.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var globals = require('syracuse-tablet/html/js/helpers/globals');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var authoringSData = require('syracuse-tablet/html/js/authoring/authoringSData');
var modal = require('syracuse-tablet/html/js/ui/modal');
var authProps = require('syracuse-tablet/html/js/authoring/properties');

var _lastHoverElement;
var _onSelectionChange;

exports.enableAuthCss = function(onSelectionChange) {
	$(".s-m-control").addClass("s-m-auth");
	$(".s-m-layout").addClass("s-m-auth");
	authProps.elementsNoAuthoring.forEach(function(e) {
		$(e).removeClass("s-m-auth");
	});
	$(".s-m-auth").on("click", _OnSelectionChange);

	$(".s-m-auth").mouseover(function(e) {
		var o = $(e.currentTarget);
		if (_lastHoverElement) {
			_lastHoverElement.removeClass("s-m-auth-hover");
		}
		o.addClass("s-m-auth-hover");
		_lastHoverElement = o;
		return false;
	});

	_onSelectionChange = onSelectionChange;
};

exports.disableAuthCss = function() {
	$(".s-m-auth").off("click", _OnSelectionChange);
	$(".s-m-auth").removeClass("s-m-auth");
	_onSelectionChange = null;
};

function _OnSelectionChange() {
	$(".s-m-auth-hover").removeClass("s-m-auth-hover");
	$(".s-m-auth-selected").removeClass("s-m-auth-selected");

	var $$s = $(this);
	$$s.addClass("s-m-auth-selected");

	var sel = exports.getSelection();
	var selData = exports.getSelectionData(sel);
	if (!selData || !selData.articleSelection) {
		$$s.removeClass("s-m-auth-selected");
	}
	if (_onSelectionChange) {
		_onSelectionChange(sel);
	}

	return false;
}

exports.getSelection = function() {
	var lap = $(".s-m-auth-selected");
	if (lap.length !== 1) {
		return null;
	}
	var sel = {
		$$elmt: lap,
		layoutPath: lap.attr("data-layout-path")
	};
	return sel;
};

exports.getSelectionData = function(sel) {
	if (!sel || !sel.layoutPath) {
		return null;
	}
	var app = globals.getApplication();
	var article = $.extend(true, {}, app.currentPage.article);
	var node = {
		"$items": [article]
	};

	var elmts = sel.layoutPath.split("-");
	while (elmts.length > 0) {
		var idx = +elmts.shift();
		var children;
		if (node && node.$items) {
			children = node.$items;
		} else if (node.$article) {
			children = [node.$article];
		}
		node = children && children[idx];
	}
	var elmt;

	return {
		articlePage: article,
		articleSelection: node,
		structElement: elmt
	};
};

exports.saveAuthoring = function(page, opts) {

	var segs = page.split(".");

	var pageData = {
		application: segs[0], // x3
		contract: segs[1], // erp
		representation: segs[2], // ACTIV
		facet: segs[3], // $query

		roles: [],
		users: [],
		endpoints: [],

		saveAsOption: opts.saveAs,

		variantCode: opts.code,
		variantTitle: opts.title,
		variantDescription: opts.description,

		article: opts.article
	};

	return authoringSData.getPageVariants({
		application: segs[0],
		contract: segs[1],
		representation: segs[2],
		facet: segs[3]
	})
		.then(function(variants) {
			variants && variants.some(function(v) {
				var match = v.title === opts.title && v.code === opts.code && v.description === opts.description;

				// Did type change? Then create new variant
				if (opts.saveAs === "factory_variant" && v.isFactory !== true) {
					match = false;
				} else if (opts.saveAs === "global_variant" && (v.isFactory === true || v.personalCopy === true)) {
					match = false;
				} else if (opts.saveAs === "personal_copy" && v.personalCopy !== true) {
					match = false;
				}

				// If all matches, it's a save and not a save as
				if (match) {
					pageData.variant = v.$uuid;
				}
			});
		})
		.then(function() {
			return authoringSData.savePageDefinition(pageData);
		})
		.then(function(data) {
			modal.info(locale.text("auth.panel.save_ok_title"), locale.text("auth.panel.save_ok_text"));
		})
		.fail(function(e) {
			modal.info(locale.text("auth.panel.save_error_title"), locale.text("auth.panel.save_error_text") + "\n" + JSON.stringify(e));
		});
};

exports.validateJSON = function(jsonText) {
	var obj;
	try {
		obj = JSON.parse(jsonText);

		// TODO: Logically check article content

		return obj;
	} catch (e) {
		modal.info(locale.text("auth.panel.save_error_title"), locale.text("auth.panel.save_error_json"));
	}
};

exports.toSortedJSON = function(obj) {
	var toJSON = function() {
		var self = this;
		var that = {};
		var $items;

		Object.keys(self).forEach(function(key) {
			if (key !== "$items") {
				that[key] = self[key];
			} else {
				$items = self[key];
			}
		});
		if ($items) {
			that.$items = $items;
		}
		return that;
	};
	// Step one: Hook property ordering
	JSON.stringify(obj,
		function replacer(key, value) {
			if ($.isPlainObject(value)) {
				value.toJSON = toJSON;
			}
			return value;
		});
	// Build json
	var str = JSON.stringify(obj, null, " ");
	return str;
};
