// syracuse-tablet/html/js/authoring/authoringDrag.js
"use strict";

var authHelpers = require('syracuse-tablet/html/js/authoring/authoringHelpers');
var ctrlFactory = require('syracuse-tablet/html/js/controls/ctrlFactory');
var notifications = require('syracuse-tablet/html/js/helpers/notifications');

/*
	Drag and drop helpers
*/

/*
 * Gets the control/drag element a dragged DOM element belongs to
 * (The DOM element on drag start can be a child of the actual draggable, this is why we need to go up the tree)
 */
exports.getClosestDragElement = function($$source) {
	var $$draggable = $$source.closest(".ui-draggable");
	return $$draggable;
};


exports.dropDrop = function($$drop, $$helper, $$draggable) {
	if (!_isDropAllowed($$drop, $$draggable, $$helper)) {
		return;
	}

	var data = _getDropData($$drop, $$draggable, $$helper);
	if (data.action === "add_layout") {
		_addLayout(data);
	}
};

exports.dropAccept = function($$drop, $$draggable) {
	return _isDropAllowed($$drop, $$draggable);
};

exports.dropOver = function($$drop, $$helper, $$target) {

};

function _isDropAllowed($$drop, $$draggable, $$helper) {
	var data = _getDropData($$drop, $$draggable, $$helper);
	if (!data.toType) {
		return false;
	}
	if (data.action === "add_layout") {
		if (data.bind == "stack") {
			if (data.toType !== "stack" && data.toType !== "cell" && data.toType !== "tile") {
				return false;
			}
		} else if (data.bind == "group") {
			if (data.toType !== "stack" && data.toType !== "hub") {
				return false;
			}
		} else if (data.bind == "tile") {
			if (data.toType !== "row" && data.toType !== "hubGroup") {
				return false;
			}
		}
	}
	return true;
}

function _getDropData($$drop, $$draggable, $$helper) {
	var action;
	var fromId;
	var toId;
	var bind;
	var parent;

	action = $$draggable.attr("data-auth-drag-action");
	switch (action) {
		case "move":
			fromId = $$draggable.attr("data-ctrl-id");
			break;
		case "add_bind":
			bind = $$draggable.attr("data-auth-bind");
			parent = $$draggable.attr("data-auth-bind-parent");
			break;
		case "add_layout":
			bind = $$draggable.attr("data-auth-bind");
			break;
		default:
			// drop from page
			action = "move";
			fromId = $$draggable.attr("id");
			break;
	}

	var $$ctrl;
	toId = $$drop.attr("data-ctrl-id"); // drop inside tree
	if (toId && toId.length > 0) {
		$$ctrl = $("#" + toId);
	} else {
		toId = $$drop.attr("id"); // Drop inside page
		$$ctrl = $$drop;
	}

	// drop control or layout
	var ctrl = authHelpers.findControl($$ctrl);
	return {
		action: action, // move, add_bind, add_layout
		bind: bind, // $bind or name of layout in case of add*
		parent: parent, // parent $bind (e.g. vignette name)
		fromId: fromId, // ctrlId (when dragging existing ctrl)
		toId: toId, // ctrlId of drop target
		toType: ctrl && ctrl.$type, // Type of drop target
		toCtrl: ctrl
	};
}

function _addLayout(data) {
	var layoutType = "stack";

	switch (data.bind) {
		case "group":
			if (data.toType == "hub") {
				layoutType = "hubGroup";
			} else {
				layoutType = "row";
			}
			break;
		case "tile":
			if (data.toType == "hubGroup") {
				layoutType = "tile";
			} else {
				layoutType = "cell";
			}
			break;
		case "stack":
			layoutType = "stack";
			break;
	}

	var article = {
		$layoutType: layoutType
	};

	// Create new layout node
	ctrlFactory.createLayout(data.toCtrl.controller, data.toCtrl, article, {});
	notifications.publish(["sm.auth.change.item.prop.ui"]);
}
