// syracuse-tablet/html/js/authoring/authoringPage.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var notifications = require('syracuse-tablet/html/js/helpers/notifications');

var $$lastHoverElement;

//Controls and layout classes that cannot be selected during authoring
var _elementsNoAuthoring = [".s-m-hub", ".s-m-array .s-m-control", ".s-m-array .s-m-layout"];

exports.AuthPage = utils.defineClass(
	function AuthPage() {
		var self = this;
		notifications.subscribe(self, ["sm.auth.hover.in"]);
		notifications.subscribe(self, ["sm.auth.hover.out"]);

		notifications.subscribe(self, ["sm.auth.set.selection"]);
		notifications.subscribe(self, ["sm.auth.remove.selection"]);

		self.$$page = $("#s-m-app-id");
	}, null, {
		/*
		 * Enable hover and click events in page
		 */
		enableAuthCss: function() {
			var self = this;
			$(".s-m-control").addClass("s-m-auth");
			$(".s-m-layout").addClass("s-m-auth");

			_elementsNoAuthoring.forEach(function(e) {
				$(e).removeClass("s-m-auth");
			});

			$(".s-m-auth", self.$$page).on("click", function(e) {
				var $$o = $(e.currentTarget);

				var $$old = $(".s-m-auth-selected", self.$$page);
				$$old.removeClass("s-m-auth-selected");
				var path = $$old.attr("data-layout-path");
				notifications.publish(["sm.auth.remove.selection"], path);

				$$o.addClass("s-m-auth-selected");
				path = $$o.attr("data-layout-path");
				notifications.publish(["sm.auth.set.selection"], path);
				return false;
			});

			$(".s-m-auth").mouseover(function(e) {
				var $$o = $(e.currentTarget);
				if ($$lastHoverElement) {
					$$lastHoverElement.removeClass("s-m-auth-hover");
					notifications.publish(["sm.auth.hover.out"], $$lastHoverElement.attr("data-layout-path"));
				}
				notifications.publish(["sm.auth.hover.in"], $$o.attr("data-layout-path"));
				$$o.addClass("s-m-auth-hover");
				$$lastHoverElement = $$o;
				return false;
			});

			$(".s-m-auth").mouseout(function(e) {
				var $$o = $(e.currentTarget);
				$$o.removeClass("s-m-auth-hover");
				notifications.publish(["sm.auth.hover.out"], $$o.attr("data-layout-path"));
			});
		},

		/*
		 * Disable all events in page that were used for authoring interaction
		 */
		disableAuthCss: function() {
			var self = this;
			$(".s-m-auth", self.$$page).removeClass("s-m-auth");
		},

		notifAuthHoverIn: function(path) {
			var self = this;
			var $$pageItem = self._getItemByPath(path);
			$$pageItem.addClass("s-m-auth-hover");
		},

		notifAuthHoverOut: function(path) {
			var self = this;
			var $$pageItem = self._getItemByPath(path);
			$$pageItem.removeClass("s-m-auth-hover");
		},

		notifAuthSetSelection: function(path) {
			var self = this;
			var $$item = self._getItemByPath(path);
			$$item.addClass("s-m-auth-selected");
		},

		notifAuthRemoveSelection: function(path) {
			var self = this;
			var $$item = self._getItemByPath(path);
			$$item.removeClass("s-m-auth-selected");
		},

		_getItemByPath: function(path) {
			var self = this;
			return $("[data-layout-path=" + path + "]", self.$$page);
		},

		getCurrentSelectionPath: function() {
			var self = this;
			var $$item = $(".s-m-auth-selected", self.$$page);
			return $$item.attr("data-layout-path");
		}
	});
