// syracuse-tablet/html/js/authoring/authoringPropertiesPages.js
"use strict";

var authoringActions = require('syracuse-tablet/html/js/authoring/page/authoringActions');
var authoringChartDetail = require('syracuse-tablet/html/js/authoring/page/authoringChartDetail');
var authPropsGeneral = require('syracuse-tablet/html/js/authoring/authoringPropertiesGeneral');
var authComponents = require('syracuse-tablet/html/js/authoring/authoringComponents');
var authUtils = require('syracuse-tablet/html/js/authoring/authoringPropertiesGeneral');
var jsutils = require('syracuse-tablet/html/js/helpers/jsutils');
var locale = require('syracuse-tablet/html/js/helpers/locale');

var _properties = {
	"variants": function(controller, sel, data) {
		this.key = "$variant";
		this.views = data.page.pageData.page.$views;
		if (!this.views || this.views.length <= 1) {
			return;
		};
		var titles = [];
		var currentSelection;
		this.views.forEach(function(view, idx) {
			titles.push({
				"title": view.$title,
				"value": view.uuid || view.$url,
				"selected": view.$selected
			});
			if (view.$selected) currentSelection = view.uuid || view.$url;
		});
		this.applyMarkup = function($$parent, prop, sel, data) {
			var subsection = this._addSubSection($$parent);
			var opts = {
				css: "variants-list",
				value: currentSelection,
				options: titles,
				onChange: jsutils.bindFn(this.valueChange, this, sel, data, prop),
				dataSize: 20
			};
			this.viewSelect = authComponents.newComponent("select", opts);
			this.viewSelect.createMarkup(subsection.$$content);
			if (data.articleSelection && data.articleSelection.$variant) {
				this.viewSelect.setValue(data.articleSelection.$variant);
			};
		};

		this._addSubSection = function($$parent) {
			var title = locale.text("auth.page.views");
			var id = "pageViewsList";
			return authoringActions.addSubSection(title, id, $$parent);
		};

		this.valueChange = function(event, viewSelect, sel, data, prop) {
			var uuid = viewSelect.getValue();
			var self = this;
			var variants = data && data.page && data.page.variants;
			if (!variants) return;

			var variantSelected = $.smFind(variants, function(variant) {
				if (uuid.indexOf(variant.$uuid) > -1) {
					return true;
				};
				return false;
			});
			if (variantSelected) {
				this.views.forEach(function(view) {
					view.$selected = false;
					if (view.$url === uuid || view.uuid == uuid) {
						view.$selected = true;
					}
				});
				authPropsGeneral.notifyAuthVariantChange(JSON.parse(variantSelected.pageData.content).$article);
			}
		};
	},
	/*
	 * display stack as navigation bar
	 */
	"stackbar": function(controller, sel, data) {
		if (!data.page.controller.isStackBarAvailable()) return null;
		this.key = "$stackbar";
		this["default"] = data.articleSelection && data.articleSelection.$stackbar || false;
		this.values = [{
			"value": true
		}, {
			"value": false
		}];
		this.name = "stackbar";
		this.createMarkup = authUtils.createMarkupRadio
	}
};

var _modules = [authoringActions, authoringChartDetail];

exports.initModuleProperties = function(dest) {
	_modules.forEach(function(module) {
		module.initModuleProperties(dest);
	});
};

exports.getModuleProperties = function($$elmt, controller, props, sel, data) {
	var page = data.page ? data.page : data.control && data.control.isVignette() ? data.control.getTopPage() : null;
	if (page == null) return;
	if (data.control == null) {
		props.push(_properties.variants);
	} else {
		data.page = page;
	}
	if (!page.isCardDetail() && !page.isVignette() && !page.isDashboard()) {
		if (page.isFacet("$details", "$edit", "$create")) props.push(_properties.stackbar);
		authoringActions.getModuleProperties($$elmt, controller, props, sel, data);
	}
};
