// syracuse-tablet/html/js/authoring/authoringPropertiesGeneral.js
"use strict";
var authHtml = require('syracuse-tablet/html/js/authoring/authoringHtml');
var authVal = require('syracuse-tablet/html/js/authoring/authoringVal');
var notifications = require('syracuse-tablet/html/js/helpers/notifications');
var modal = require('syracuse-tablet/html/js/ui/modals/modal');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var jsutils = require('syracuse-tablet/html/js/helpers/jsutils');
var globals = require('syracuse-tablet/html/js/helpers/globals');
var authComponents = require('syracuse-tablet/html/js/authoring/authoringComponents');
var sdataCommonResources = require('syracuse-tablet/html/js/sdata/sdataCommonResources');
var modalTranslationPage = require('syracuse-tablet/html/js/ui/modals/modalTranslationPage');

var _colorPalette = [{
	title: "primecolor",
	items: ["sagegreen", "green", "blue", "darkgrey", "white", "black"]
}, {
	title: "secondcolor",
	items: ["lightgreen", "grey", "medium-grey", "orange", "purple", "medium-purple", "skyblue"]
}, {
	title: "notifcolors",
	items: ["alert", "warning"]
}];
var _cachePalette = {};
var _getColorPalette = function() {
	if (_cachePalette.colorPalette) {
		return _cachePalette.colorPalette;
	}
	_cachePalette.colorPalette = [];
	var _scan = function(items, array) {
		array = array || [];
		items.forEach(function(item) {
			if (item.items) {
				var items = _scan(item.items);
				array.push({
					title: item.title ? locale.text("colorsPalette." + item.title) : null,
					items: items
				});
			} else if (typeof item === "string") {
				var id = item.toLowerCase();
				array.push({
					id: id,
					name: locale.textNoFail("colorsPalette.colors." + id) || item.smCapitalize(),
					css: "s-m-color-" + id
				});
			}
		});
		return array;
	};
	return _scan(_colorPalette, _cachePalette.colorPalette);
};
/**
 * Returns a standard value/label list
 */
var _getColorValues = function() {
	if (_cachePalette.colorValues) {
		return _cachePalette.colorValues;
	}
	_cachePalette.colorValues = [];
	var _scan = function(items) {
		items.forEach(function(item) {
			if (item.items) {
				_scan(item.items);
			} else if (typeof item === "string") {
				_cachePalette.colorValues.push({
					value: item.toLowerCase(),
					label: locale.textNoFail("colorsPalette.colors." + item) || item.smCapitalize()
				});
			}
		});
	};
	_scan(_colorPalette);
	return _cachePalette.colorValues;
};
/**
 * Returns info  for color id
 */
var _getColorInfo = function(id) {
	if (_cachePalette.colorInfos) {
		return _cachePalette.colorInfos[id];
	}
	_cachePalette.colorInfos = {};
	_getColorValues().forEach(function(color) {
		_cachePalette.colorInfos[color.value] = {
			id: color.value,
			name: color.label,
			css: "s-m-color-" + color.value
		};
	});
	return _cachePalette.colorInfos[id];
};
var _openInNewTab = function(tabid, jsonString, title) {
	if (!jsonString || jsonString.trim().length === 0) return;
	var myWindow = window.open("about:blank", tabid, undefined, true);
	setTimeout(function() {
		myWindow.document.write("<html><body><pre>" + (title ? title + "\n" : "") + locale.text("auth.panel.label_date") + ": " + (new Date()).toLocaleString() + "</pre><pre>" + jsonString + "</pre></body></html>");
	});
};

var _properties = {
	/*
	 * Dummy to add a separator
	 */
	"SEPARATOR": {
		isSeparatror: true
	},

	/*
	 * Full JSON article of selection for experts
	 */
	"JSON": {
		"key": "$JSON",
		createMarkup: _createMarkupJSON,
		type: "expert"
	},
	/*
	 * Read only display of prototype
	 */
	"PROTOTYPE": {
		"key": "$PROTOYPE",
		createMarkup: _createMarkupPrototype,
		type: "expert"
	},
	"translation": function(controller, sel, data) {
		if (!data.articleSelection.$localization || $.isEmptyObject(data.articleSelection.$localization[globals.getUserCtx().$lang])) return null;
		this.name = "translation";
		this.type = "expert";
		this.createMarkup = function($$parent, prop, sel, data) {
			var self = this;
			sdataCommonResources.queryLocalePreferences().then(function(ls) {
				var selectOpts = [];
				ls.forEach(function(v) {
					selectOpts.push({
						'title': v.description,
						'value': v.code
					});
				});
				var opts = {
					options: selectOpts,
					//onChange: self.onChange.bind(self, $$parent)
					onChange: self.onChange.bind(self, $$parent, data)

				};
				self.langList = authComponents.newComponent("select", opts, "selectLang");
				self.langList.createMarkup($$parent);
			});

		};
		this.onChange = function($$parent, data, event, componentSelect) {
			var context = {};
			context.$localization = data.articleSelection.$localization;
			context.currentLang = {
				'value': globals.getUserCtx().$lang,
				'title': getLangDescription(globals.getUserCtx().$lang, componentSelect.options.options)
			};
			context.translateLang = {
				'value': componentSelect.getValue(),
				'title': getLangDescription(componentSelect.getValue(), componentSelect.options.options)
			};
			this.buildPanel($$parent, data, context);
		};
		this.buildPanel = function($$parent, data, context) {
			if (this.$$panel == undefined) {
				var html = authHtml.execute("authPanelButton", {
					label: locale.text("auth.translation.openPage"),
					css: "s-m-auth-translation-open"
				});
				this.$$panel = $(html);
				this.$$panel.attr("data-action", "openTranslation");
				$$parent.append(this.$$panel);
			} else {
				this.$$panel.off("click");
			}
			var eventData = {};
			eventData.$$parent = $$parent;
			eventData.data = data;
			eventData.context = context;
			eventData.context.pageTitle = locale.text("auth.translation.page"),
			this.$$panel.click(eventData, function() {
				var modal = new modalTranslationPage.Modal(eventData.context);
				modal.show().then(function(result) {
					if (result) {
						eventData.data.articleSelection.$localization = result;
						_notifyChangePropUI(eventData.$$parent, null, eventData.data);
					}
				}).fail(function(e) {
					globals.getModal().error(e);
				});
			});
		};
	}
};

function _createMarkupJSON($$parent, prop, sel, data) {
	// This is only displayed if the full page is selected, so we always show the article of the page and not the article
	// of the selected UI compontent (UI component is always the page)
	var value = authVal.toSortedJSON(data.articlePage);
	var html = authHtml.execute("authPanelPropertyJSON", {
		id: "auth_opt_" + prop.name,
		value: value,
		label_apply: locale.text("auth.panel.label_apply"),
		label_display: locale.text("auth.panel.label_display")
	});
	var $$panel = $(html);
	var title = sel.$$elmt.text();
	var _getJson = function(validate) {
		var val = $("textarea", $$panel).val();
		var articleParsed;
		if (validate) {
			articleParsed = authVal.validateJSON(true, $("textarea", $$panel).val(), _properties);
		} else {
			try {
				articleParsed = JSON.parse(val);
			} catch (e) {
				return jsutils.convertToDiagnoses(e);
			}
		}
		return articleParsed ? authVal.toSortedJSON(articleParsed) : null;
	};
	$("a[data-action]", $$panel).click(function() {
		var act = $(this).attr("data-action");
		try {
			if (act === "jsonApply") {
				// Full page article exchange
				var articleString = _getJson(true);
				if (articleString) {
					notifications.publish(["sm.auth.change.prop.ui"], articleString);
				}
			} else if (act === "openInNewTab") {
				var title = globals.getApplication().currentPage.prototype;
				if (title) title = title.data("$title");
				title = title && title.length > 0 ? locale.text("auth.panel.label_page") + ": " + title : "";
				_openInNewTab("sageMobileJsonAuthoring", _getJson(false), title);
			}
		} catch (e) {
			console.log(e.stack);
			throw (e);
		}
	});
	$$parent.append($$panel);
}

function _createMarkupPrototype($$parent, prop, sel, data) {
	var json;
	var isPage = sel.$$elmt && sel.$$elmt.is(".s-m-page");
	var title;
	if (data.control) {
		json = data.control.prototype && data.control.prototype.json;
		title = data.control.$bind || data.control.id;
	} else if (isPage) {
		var controller = sel.$$elmt.smPageController();
		if (controller) {
			title = controller.prototype.data("$title");
			json = controller.prototype && controller.prototype.json;
		}
	}
	var value;
	if (json) {
		value = JSON.stringify(json, null, " ");
	} else {
		value = locale.text("auth.PROTOTYPE_NONE");
	}
	var html = authHtml.execute("authPanelPropertyPROTOTYPE", {
		id: "auth_opt_" + prop.name,
		value: value,
		label_display: locale.text("auth.panel.label_display")
	});
	var $$panel = $(html);
	$$parent.append($$panel);
	if (json && !$.isEmptyObject(json)) {
		$("a[data-action]", $$panel).click(function() {
			var act = $(this).attr("data-action");
			if (act === "openInNewTab") {
				title = locale.text("auth.panel.label_" + (isPage ? "page" : "control")) + ": " + title;
				_openInNewTab("sageMobilePrototype", authVal.toSortedJSON(json), title);
			}
		});
	}
}
var _initProperty = function(p) {
	var pre = "auth";
	var lk = pre + "." + p.name;
	if (p.label == null) {
		p.label = locale.text(lk);
	}
	if (p.values) {
		p.values.forEach(function(v, idx) {
			if ($.isPlainObject(v)) {
				var vk = lk + "." + ("" + ((v.value != null ? v.value : v.key)));
				if (v.label == null) {
					v.label = locale.text(vk);
				}
			} else {
				p.values[idx] = {
					label: v + "",
					valu: v
				};
			}
		});
	}
};
var _initProperties = function(properties, dest) {
	var pre = "auth";
	Object.keys(properties).forEach(function(prop) {
		var p = properties[prop];
		if (dest) {
			dest[prop] = p;
		}
		if (typeof p === "function") {
			return;
		}
		p.name = prop;
		_initProperty(p);
	});
};

var _err = function(title, ee) {
	modal.error(title, ee);
	return true;
};

function getLangDescription(lang, langList) {
	var desc;
	langList.forEach(function(v) {
		if (v.value === lang) {
			desc = v.title;
		}
	});
	return desc;
}

function _getValueOrDefault(prop, sel, data) {
	var value = data.articleSelection && data.articleSelection[prop.key];
	value = (value != null) ? value : prop["default"];
	return value;
}

function _setValue(prop, sel, data, value) {
	data.articleSelection[prop.key] = value;

	// To flag control that has been changed
	data.articleSelection["$isDirty"] = data.articleSelection["$isDirty"] || {};
	data.articleSelection["$isDirty"][prop.key] = true;
}

function _notifyChangePropUI($$parent, sel, data) {
	var $$json = $("textarea#auth_opt_JSON");
	$$json.val(authVal.toSortedJSON(data.articleSelection));
	notifications.publish(["sm.auth.change.item.prop.ui"], data);
}

function _notifyAuthVariantChange(article) {
	var articleString = authVal.toSortedJSON(article);
	$("textarea#auth_opt_JSON").val(articleString);
	notifications.publish(["sm.auth.change.prop.ui"], articleString);
}

function _createMarkupRadio($$parent, prop, sel, data) {

	var ctx = {
		group: "auth_opt_" + prop.name,
		options: []
	};

	var value = _getValueOrDefault(prop, sel, data);
	var allHidden = true;

	prop.values.forEach(function(v) {
		ctx.options.push({
			id: ctx.group + "_" + v.value,
			value: "" + v.value,
			label: v.label,
			display: (v.isHidden ? "none" : "block")
		});
		allHidden = allHidden && v.isHidden;
	});

	var html = authHtml.execute("authPanelPropertyRadio", ctx);
	var $$panel = $(html);
	$("label[data-s-m-auth-value='" + ("" + value) + "']", $$panel).eq(0).button("toggle");
	$("label", $$panel).on("click", function(e) {
		var $$ct = $(e.currentTarget);
		var value = $$ct.attr("data-s-m-auth-value");
		$$ct.eq(0).button("toggle");

		/*
		 * Important! Use boolean datatype for boolean properties!
		 */
		if (value === "true") {
			value = true;
		} else if (value === "false") {
			value = false;
		}
		_setValue(prop, sel, data, value);
		_notifyChangePropUI($$parent, sel, data);
	});
	$$parent.append($$panel);

	if (allHidden) {
		$$parent.parent(".s-m-panel-auth-prop").hide();
	} else {
		$$parent.parent(".s-m-panel-auth-prop").show();
	}
}

function _refreshMarkupRadio($$parent, prop, sel, data) {
	var allHidden = true;
	prop.values.forEach(function(v) {
		var id = "auth_opt_" + prop.name + "_" + v.value;
		var $$btn = $("#" + id, $$parent);
		var display = v.isHidden ? "none" : "block";
		$$btn.css("display", display);
		allHidden = allHidden && v.isHidden;
	});
	if (allHidden) {
		$$parent.parent(".s-m-panel-auth-prop").hide();
	} else {
		$$parent.parent(".s-m-panel-auth-prop").show();
	}
}

function _createMarkupCheckboxes($$parent, prop, sel, data) {

	var ctx = {
		group: "auth_opt_" + prop.name,
		options: []
	};

	var allHidden = true;

	var value = data.articleSelection && data.articleSelection[prop.key] || {};
	prop.values.forEach(function(v) {
		var cb = value[v.key] != null ? value[v.key] : v["default"];
		ctx.options.push({
			id: ctx.group + "_" + v.key,
			value: v.key,
			label: v.label,
			checked: cb,
			display: (v.isHidden ? "none" : "block")
		});
		allHidden = allHidden && v.isHidden;
	});

	var html = authHtml.execute("authPanelPropertyCheckbox", ctx);
	var $$panel = $(html);

	$("label", $$panel).on("click", function(e) {
		// #7840 We use an icon instead of a check box because click with input tag doesn't work fin with the label
		var $$ct = $(e.target);
		var key = $$ct.attr("data-s-m-auth-value");
		var $$cb = $("i", $$ct);
		var value = $("i", $$ct).toggleClass("fa-check-square").is(".fa-check-square");
		var v = data.articleSelection[prop.key] = data.articleSelection[prop.key] || {};
		v[key] = value;
		_notifyChangePropUI($$parent, sel, data);
	});
	$$parent.append($$panel);

	if (allHidden) {
		$$parent.parent(".s-m-panel-auth-prop").hide();
	} else {
		$$parent.parent(".s-m-panel-auth-prop").show();
	}
}

function _refreshMarkupCheckboxes($$parent, prop, sel, data) {
	var allHidden = true;
	prop.values.forEach(function(v) {
		var id = "auth_opt_" + prop.name + "_" + v.key;
		var $$btn = $("#" + id, $$parent);
		var display = v.isHidden ? "none" : "block";
		$$btn.css("display", display);
		allHidden = allHidden && v.isHidden;
	});
	if (allHidden) {
		$$parent.parent(".s-m-panel-auth-prop").hide();
	} else {
		$$parent.parent(".s-m-panel-auth-prop").show();
	}
}

function _createMarkupIconPicker($$parent, prop, sel, data) {
	var value = data.articleSelection && data.articleSelection[prop.key] || null;
	var opts = {
		palette: prop["values"] || [],
		dropUp: prop["dropUp"] || true,
		onSelectIcon: function(value) {
			_setValue(prop, sel, data, value);
			_notifyChangePropUI($$parent, sel, data);
		}
	};
	prop._paletteIcon = authComponents.newComponent("iconPalettePicker", opts);
	prop._paletteIcon.createMarkup($$parent);
	prop._paletteIcon.setValue(data.articleSelection[prop.key] || prop["default"]);
}

function _destroyMarkupIconPicker($$parent, prop, sel, data) {
	if (prop._paletteIcon) {
		prop._paletteIcon.destroyMarkup();
		prop._paletteIcon = null;
	}
}

exports.getColorValues = _getColorValues;
exports.getColorPalette = _getColorPalette;
exports.getColorInfo = _getColorInfo;

exports.properties = _properties;
exports.err = _err;
exports.createMarkupIconPicker = _createMarkupIconPicker;
exports.destroyMarkupIconPicker = _destroyMarkupIconPicker;
exports.createMarkupRadio = _createMarkupRadio;
exports.refreshMarkupRadio = _refreshMarkupRadio;
exports.createMarkupCheckboxes = _createMarkupCheckboxes;
exports.refreshMarkupCheckboxes = _refreshMarkupCheckboxes;
exports.getValueOrDefault = _getValueOrDefault;
exports.setValue = _setValue;
exports.notifyAuthVariantChange = _notifyAuthVariantChange;
exports.notifyChangePropUI = _notifyChangePropUI;
exports.initProperties = _initProperties;
exports.initProperty = _initProperty;
exports.getModuleProperties = function($$elmt, controller, props, sel, data) {
	// Always add JSON editting and proto display
	props.push(_properties.SEPARATOR);
	if ($$elmt.is(".s-m-page")) {
		props.push(_properties.JSON);
		props.push(_properties.SEPARATOR);
		props.push(_properties.PROTOTYPE);
		props.push(_properties.SEPARATOR);
		props.push(_properties.translation);
	} else if (data && data.control && data.control.prototype) {
		props.push(_properties.PROTOTYPE);
	}
};
exports.initModuleProperties = function(dest) {
	exports.initProperties(_properties, dest);
};
