// syracuse-tablet/html/js/authoring/authoringHtml.js
"use strict";

/*
 * Authoring panel main layout
 */
var _templates = {
	authPanelHeader: '\
		<div class="btn-group"> \
			<div class="btn-group"> \
				<button class="btn btn-auth-header" id="s-m-auth-settings-id" type="button"><label class="glyphicon glyphicon-cog"></label>{{label_settings}}</button> \
			</div> \
			\
			<div class="btn-group" data-toggle="buttons" id="s-m-auth-orientation-id"> \
				<button class="btn btn-auth-header active" id="s-m-auth-vertical-id" data-auth-orientation="portrait"> \
					<input type="radio" name="orientation" checked> \
					<label class="glyphicon glyphicon-resize-vertical"></label> \
				</button> \
				<button class="btn btn-auth-header" id="s-m-auth-horizontal-id" data-auth-orientation="landscape"> \
					<input type="radio" name="orientation"> \
					<label class="glyphicon glyphicon-resize-horizontal"></label> \
				</button> \
			</div> \
			\
			<div class="btn-group"> \
				<button class="btn btn-auth-header" id="s-m-auth-scale-id" type="button"><label class="glyphicon glyphicon-fullscreen"></button> \
			</div> \
		</div> \
		<div class="btn-group"> \
			<button class="btn btn-auth-header disabled" id="s-m-auth-remove-id" type="button"><label class="glyphicon glyphicon-trash"></label>{{label_remove}}</button> \
			<button class="btn btn-auth-header" id="s-m-auth-remove-all-id" type="button"><label class="glyphicon glyphicon-erase"></label>{{label_remove_all}}</button> \
		</div> \
		<div class="btn-group pull-right"> \
			<div class="btn-group "> \
				<button id="auth-btn-save-id"  class="btn btn-auth-header" type="button"><label class="glyphicon glyphicon-ok"></label></button> \
				<button id="auth-btn-close-id" class="btn btn-danger" type="button"><label class="glyphicon glyphicon-remove"></label></button> \
			</div> \
		</div>',

	authPanelFooter: '',

	authPanelRightStructure: '\
		<div class="s-m-site-auth-panel-content"> \
			<ul class="nav nav-pills" role="tablist"> \
				<li><a href="#authPanelControl" role="tab" data-toggle="tab">{{label_control}}</a></li> \
			</ul> \
			<div class="tab-content"> \
				<div id="authPanelControl" class="tab-pane panel-body active"> \
		 			<div class="s-m-auth-prop-list"> \
						<label class="control-label">{{label_control_title}}</label> \
				    	<p class="form-control-static" id="authPanelControlTitle"> </p> \
						<label class="control-label">{{label_control_type}}</label> \
						<p class="form-control-static" id="authPanelControlType"> </p> \
					</div> \
					<div id="authPanelControlSimple"> \
					</div> \
				</div> \
			</div> \
		</div>',

	authPanelLeftStructure: '\
		<div class="s-m-site-auth-panel-content"> \
			<ul class="nav nav-pills" role="tablist"> \
				<li><a href="#s-m-auth-tree-box" role="tab" data-toggle="tab">{{label_content}}</a></li> \
				<li><a href="#s-m-auth-insert-box" role="tab" data-toggle="tab">{{label_insert}}</a></li> \
			</ul> \
			<div class="tab-content"> \
				<div id="s-m-auth-tree-box" class="tab-pane active"> \
					<div class="s-m-scroll-wrapper s-m-scroll-wrapper-v">\
						<ul class="s-m-auth-tree s-m-scroll-element" id="auth-structure-root-id"></ul> \
					</div> \
				</div> \
				<div id="s-m-auth-insert-box" class="tab-pane"> \
					<div class="panel panel-primary s-m-panel-auth-prop"> \
						<div class="panel-heading" role="tab" id="auth-lp-cnt-containers-header"> \
							<a data-toggle="collapse" data-parent="#accordion" href="#auth-lp-cnt-containers-clpse" \
								aria-expanded="true" aria-controls="auth-lp-cnt-containers-clpse" \
								<h4 class="panel-title"> \
									{{label_containers}}\
								</h4> \
							</a> \
						</div> \
						<div id="auth-lp-cnt-containers-clpse" class="panel-collapse collapse in" \
							role="tabpanel" aria-labelledby="auth-lp-cnt-containers-header"> \
							<div class="s-m-auth-btn-container s-m-draggable fa fa-th" data-auth-bind="group" data-auth-drag-action="add_layout"></div> \
							<div class="s-m-auth-btn-container s-m-draggable fa fa-square-o" data-auth-bind="tile" data-auth-drag-action="add_layout"></div> \
							<div class="s-m-auth-btn-container s-m-draggable fa fa-bars" data-auth-bind="stack" data-auth-drag-action="add_layout"></div> \
						</div> \
					</div> \
					<div class="panel panel-primary s-m-panel-auth-prop"> \
						<div class="panel-heading" role="tab" id="auth-lp-cnt-props-header"> \
							<a data-toggle="collapse" data-parent="#accordion" href="#auth-lp-cnt-props-clpse" \
								aria-expanded="true" aria-controls="auth-lp-cnt-props-clpse">\
								<h4 class="panel-title"> \
									{{label_fields}}\
								</h4> \
							</a> \
						</div> \
						<div id="auth-lp-cnt-props-clpse" class="panel-collapse collapse in" role="tabpanel" aria-labelledby="auth-lp-cnt-props-header"> \
						</div> \
					</div> \
				</div> \
			</div> \
		</div>',

	authPanelPropertyGroup: ' \
		<div class="panel panel-primary s-m-panel-auth-prop s-m-auth-prop-type-{{type}}"> \
			<div class="panel-heading" role="tab" id="heading_{{id}}"> \
				<a data-toggle="collapse" data-parent="#accordion" href="#collapse_{{id}}" {{#if expanded}} aria-expanded="true"{{/if}} aria-controls="collapse_{{id}}" {{#unless expanded}}class="collapsed"{{/unless}}>\
					<h4 class="panel-title"> \
						{{label}}\
					</h4> \
				</a> \
			</div> \
			<div id="collapse_{{id}}" class="panel-collapse collapse {{#if expanded}}in{{/if}}" role="tabpanel" aria-labelledby="heading_{{id}}"> \
			</div> \
		</div>',
	authPanelPropertySeperator: ' \
		<div class="auth-group-seperator"></div> \
		',
	authPanelPropertyRadio: ' \
		<div class="btn-group" data-toggle="buttons" style="width: 100%;"> \
			{{#each options}} \
				<div class="btn-group" style="width: 100%; display: {{this.display}};" id="{{this.id}}"> \
					<label class="s-m-auth-btn-radio-icon s-m-auth-btn-block" id="{{this.id}}_lab" data-s-m-auth-value="{{this.value}}"> \
						<input type="radio" name="{{group}}" id="{{this.id}}">{{this.label}} \
					</label> \
				</div> \
			{{/each}} \
		</div>',

	authPanelPropertyRadioColor: ' \
		<div class="btn-group" data-toggle="buttons" style="width: 100%;"> \
			{{#each options}} \
				<div class="btn-group"> \
					<label class="auth-btn-color s-m-bg-{{color}}" id="{{this.id}}_lab" data-s-m-auth-value="{{this.value}}"> \
						<input type="radio" name="{{group}}" id="{{this.id}}"> \
					</label> \
				</div> \
			{{/each}} \
		</div>',

	authPanelPropertyInput: ' \
			<div class="form-group"> \
				<input class="form-control" id="{{id}}" value="{{value}}" /> \
			</div>',

	authPanelPropertyCellSize: ' \
		{{#each sizes}} \
			<label class="{{class}} auth-size-device-header" style="width: 100%">{{this.label}}</label> \
			<div class="btn-group" data-toggle="buttons" style="width: 100%;"> \
				{{#each ../values}} \
					<div class="btn-group"> \
						<label class="auth-btn-size-sel" id="{{this.id}}_lab" data-s-m-auth-prop="{{../this.prop}}" data-s-m-auth-value="{{this}}"> \
							<input type="radio" name="{{group}}" id="{{../this.prop}}={{this}}">{{this}} \
						</label> \
					</div> \
				{{/each}} \
			</div> \
		{{/each}}',

	authPanelPropertyGenLayout: ' \
		<div class="btn-group" style="width: 100%;"> \
			<label class="s-m-auth-btn-action" id="s-m-auth-new-layout-horizontal">{{label_horizontal}}</label> \
			<label class="s-m-auth-btn-action" id="s-m-auth-new-layout-vertical">{{label_vertical}}</label> \
		</div>',
	authPanelPropertyJSON: ' \
			<div class="form-group"> \
				<textarea class="form-control" id="{{id}}" rows="10">{{value}}</textarea> \
				<button class="btn btn-default pull-right" data-action="authJsonApply">{{label_apply}}</button> \
			</div><div style="clear: both;"></div>',
	authPanelPropertyPROTOTYPE: ' \
			<div class="form-group"> \
				<textarea readonly class="form-control" id="{{id}}" rows="10">{{value}}</textarea> \
			</div>',
	authPanelPropertyGBorderValues: '\
		<div class="form-group">\
		</div>\
	',
	authPanelPropertyGSegmentsNumber: '\
		<div class="form-group">\
			<textarea class="form-control" id="{{id}}" rows="1">{{value}}</textarea> \
			<button class="btn btn-default btn-block">{{label_apply}}</button> \
		</div>\
	'
};

var _compiled = {};

function _getCompiled(name) {
	var c = _compiled[name];
	if (c) {
		return c;
	}

	var t = _templates[name];
	c = Handlebars.compile(t);
	_compiled[name] = c;
	return c;
}

exports.execute = function(name, ctx) {
	var c = _getCompiled(name);
	return c(ctx);
};
