// syracuse-tablet/html/js/authoring/authoringSData.js
"use strict";

var sdataHttp = require('syracuse-tablet/html/js/sdata/sdataHttp');
var _DEVICE_NAME = "mobile";

/*
 * pageData: {
 *   application: "x3",
 *   contract: "erp",
 *   representation: "BPCUSTOMERM"
 *   facet: "$details"
 * }
 */
exports.getPageVariants = function(pageData) {
	return $.smResolve()
		.then(function() {
			pageData = $.extend(true, {}, {
				device: _DEVICE_NAME
			}, pageData);

			if (!pageData.application || !pageData.contract || !pageData.representation || !pageData.facet || !pageData.device) {
				return $.smReject("Not all mandatory parameters present: " + JSON.stringify(pageData));
			}

			var code = [pageData.application, pageData.contract, pageData.representation, pageData.facet, pageData.device].join(".");
			var req = {
				url: "/sdata/syracuse/collaboration/syracuse/pageDefs(code%20eq%20'" + code + "')?representation=pageDef.$details"
			};

			return sdataHttp.send(req)
				.then(function(response) {
					if (response && response.isSuccess) {
						return response.responseJSON.variants || [];
					}
					if (response.status === 404) {
						return [];
					}
					if (response && response.responseJSON && response.responseJSON.$diagnoses) {
						return $.smReject(response.responseJSON);
					}
					return $.smReject("Unknown error");
				});
		});
};

exports.deletePageVariants = function(pageData) {
	return $.smResolve()
		.then(function() {
			pageData = $.extend(true, {}, {
				device: _DEVICE_NAME
			}, pageData);

			if (!pageData.application || !pageData.contract || !pageData.representation || !pageData.facet || !pageData.device) {
				return $.smReject("Not all mandatory parameters present: " + JSON.stringify(pageData));
			}

			var code = [pageData.application, pageData.contract, pageData.representation, pageData.facet, pageData.device].join(".");
			var req = {
				url: "/sdata/syracuse/collaboration/syracuse/pageDefs(code%20eq%20'" + code + "')?representation=pageDef.$edit",
				method: "DELETE"
			};

			return sdataHttp.send(req)
				.then(function(response) {
					if (response && response.isSuccess) {
						return;
					}
					if (response && response.responseJSON && response.responseJSON.$diagnoses) {
						return $.smReject(response.responseJSON);
					}
					return $.smReject("Unknown error");
				});
		});
};

exports.createPageDefinitionWC = function(opts) {
	var params = {
		"pageContext": [opts.application, opts.contract, opts.representation, opts.facet].join("."),
		"device": _DEVICE_NAME,
		"representation": "pageAuth.$edit"
	};

	var url = "/sdata/syracuse/collaboration/syracuse/pageAuths" +
		(opts.variant != null ? ("('" + opts.variant + "')") : "/$template") +
		"/$workingCopies?" +
		Object.keys(params).map(function(key) {
			return key + "=" + params[key];
		}).join("&");

	var req = {
		url: url,
		method: "POST"
	};

	return sdataHttp.send(req)
		.then(function(response) {
			if (response && response.isSuccess) {
				return response.responseJSON;
			}
			if (response && response.responseJSON && response.responseJSON.$diagnoses) {
				return $.smReject(response.responseJSON);
			}
			return $.smReject("Unknown error");
		});
};
exports.savePageDefinitionWC = function(opts, wc) {
	var parameters = {
		"roles": opts.roles,
		"users": opts.users,
		"endpoints": opts.endpoints,
		"saveAsOption": opts.saveAsOption,
		"isFactory": false,
		"isModelRepresentation": false,
		"personalCopy": false,
		"variantCode": opts.variantCode,
		"variantTitle": opts.variantTitle,
		"variantDescription": opts.variantDescription
	};
	switch (opts.saveAsOption) {
		case "factory_variant":
			parameters.isFactory = true;
			break;
		case "personal_copy":
			parameters.personalCopy = true;
			break;
		case "shared_copy":
			break;
		case "global_variant":
			break;
	}

	wc.$actions = {};
	var save;
	if (opts.variant) {
		save = wc.$actions["$save"] = {};
		delete parameters.saveAsOption;
	} else {
		save = wc.$actions["saveAs"] = {};
	}
	save.$isRequested = true;
	save.$parameters = parameters;

	var req = {
		url: wc.$url,
		method: "PUT",
		send: wc
	};
	return sdataHttp.send(req)
		.then(function(response) {
			if (response && response.isSuccess) {
				return response.responseJSON;
			}
			if (response && response.responseJSON && response.responseJSON.$diagnoses) {
				return $.smReject(response.responseJSON);
			}
			return $.smReject("Unknown error");
		});
};

exports.savePageDefinition = function(opts) {
	return exports.createPageDefinitionWC(opts)
		.then(function(wc) {
			var send = $.extend(true, {}, wc);
			send.content = send.content || {};
			send.content.$article = opts.article;

			return exports.savePageDefinitionWC(opts, send)
				.then(function(wcSaved) {
					var data = $.extend(true, {}, wc, wcSaved);
					var diags = [];
					_extractDiag(data, diags);
					return {
						data: data,
						diags: diags
					};
				})
				.fail(function(e) {
					return $.smReject({
						data: null,
						diags: [{
							$severity: "error",
							$message: "" + e
						}]
					});
				});
		});
};

function _extractDiag(data, diags) {
	Object.keys(data).forEach(function(key) {
		var value = data[key];
		if (key === "$diagnoses") {
			value.forEach(function(d) {
				diags.push(d);
			});
		} else {
			if (value && typeof value === "object") {
				_extractDiag(value, diags);
			}
		}
	});
}
