// syracuse-tablet/html/js/authoring/authoringCommons.js
"use strict";

var modules = require('syracuse-tablet/html/js/common/modules');

exports.getCurrentUserLang = function() {
	var up = modules.get("appController").App.getCurrentUserProfile();
	return up.selectedLocale.code;
}

exports.getCurrentPage = function() {
	var navStack = modules.get("navStack").getMainStack();
	var currentPage = navStack && navStack.getTopPage();
	return currentPage;
}

exports.getCurrentUserUuid = function() {
	var up = modules.get("appController").App.getCurrentUserProfile();
	return up.user.$uuid;
}

exports.queryLocalePreferences = function() {
	return modules.get("ajax").request("/sdata/syracuse/collaboration/syracuse/localePreferences?representation=localePreference.$bulk")
		.then(function(result) {
			return result.data && result.data.$resources || [];
		});
}

exports.getSecurityProfile = function() {
	var up = modules.get("appController").App.getCurrentUserProfile();
	return up && up.securityProfile;
}

exports.isNonSaveablePage = function() {
	var page = exports.getCurrentPage();
	var $url = page && page.pageData && page.pageData.$url;
	if ($url && $url.indexOf("/$client/") > -1) {
		return true;
	}
	return false;
}
