// syracuse-tablet/html/js/authoring/authoringHeader.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var jsutils = require('syracuse-tablet/html/js/helpers/jsutils');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var notifications = require('syracuse-tablet/html/js/helpers/notifications');
var authHtml = require('syracuse-tablet/html/js/authoring/authoringHtml');
var globals = require('syracuse-tablet/html/js/helpers/globals');
var modalChooseDevice = require('syracuse-tablet/html/js/ui/modals/modalChooseDevice');
var authHelpers = require('syracuse-tablet/html/js/authoring/authoringHelpers');
var modalNewLayout = require('syracuse-tablet/html/js/ui/modals/modalNewLayout');

exports.Panel = utils.defineClass(function AuthoringPanelHeader() {
	var self = this;
	self.$$panel = null;
}, null, {

	show: function() {
		var self = this;
		self._ensurePanel();
		self._subscribe();
		self.$$panel.show();
		self._initDefaults();
	},
	hide: function() {
		var self = this;
		self._ensurePanel();
		self.$$panel.hide();
		self._unscribe();
	},

	_ensurePanel: function() {
		var self = this;
		if (this.$$panel) {
			return;
		}
		var ctx = {
			label_settings: locale.text("auth.device.settings"),
			label_save: locale.text("auth.panel.label_save"),
			label_remove: locale.text("auth.panel.label_remove"),
			label_remove_container: locale.text("auth.panel.label_remove_container"),
			label_remove_all: locale.text("auth.panel.label_remove_all"),
			label_new_layout: locale.text("auth.panel.label_new_layout"),
			label_preview_portrait: locale.text("auth.panel.label_preview_portrait"),
			label_preview_landscape: locale.text("auth.panel.label_preview_landscape"),
			label_scale_to_fit: locale.text("auth.panel.label_scale_to_fit"),
			label_undo: locale.text("auth.label_undo"),
			label_undo_all: locale.text("auth.label_undo_all"),
		};

		self.$$panel = $("#s-m-auth-panel-header-id");
		self.$$panel.append($(authHtml.execute("authPanelHeader", ctx)));
		self._getDomObjects();
		self._initDomEvents();
	},

	_getDomObjects: function() {
		var self = this;
		self.$$orientation = $("#s-m-auth-orientation-id", self.$$panel);
		self.$$removeItem = $("#s-m-auth-remove-id", self.$$panel);
		self.$$removeCnt = $("#s-m-auth-remove-cnt-id", self.$$panel);
		self.$$removeAllItems = $("#s-m-auth-remove-all-id", self.$$panel);
		self.$$newLayout = $("#s-m-auth-new-layout-id", self.$$panel);
		self.$$undo = $("#auth-btn-undo-id", self.$$panel);
		self.$$undoAll = $("#auth-btn-undo-all-id", self.$$panel);
	},

	_initDomEvents: function() {
		var self = this;

		// preview options
		$("#s-m-auth-settings-id").on("click", self._onChooseDevice.bind(self));
		$("#s-m-auth-scale-id").on("click", self._onScaleDevice.bind(self));
		$("#s-m-auth-vertical-id").on("click", self._onOrientation.bind(self));
		$("#s-m-auth-horizontal-id").on("click", self._onOrientation.bind(self));
		// Delete
		self.$$removeItem.on("click", self._onRemoveItem.bind(self));
		self.$$removeCnt.on("click", self._onRemoveCnt.bind(self));
		self.$$removeAllItems.on("click", self._onRemoveAllItems.bind(self));
		self.$$newLayout.on("click", self._onNewLayout.bind(self));
		self.$$undo.on("click", self._onUndo.bind(self));
		self.$$undoAll.on("click", self._onUndoAll.bind(self));
		// save/close
		$("#auth-btn-save-id").on("click", self._onSave.bind(self));
		$("#auth-btn-close-id").on("click", self._onClose.bind(self));
	},

	_initDefaults: function() {
		var self = this;
	},

	_subscribe: function() {
		var self = this;
		notifications.subscribe(self, ["sm.main.layout.changed"]);
		notifications.subscribe(self, ["sm.auth.set.selection"]);
		notifications.subscribe(self, ["sm.auth.change.prop.ui"]);
		notifications.subscribe(self, ["sm.auth.history.change"]);
	},

	_unscribe: function() {
		var self = this;
		notifications.unsubscribe(self, ["sm.main.layout.changed"]);
		notifications.unsubscribe(self, ["sm.auth.set.selection"]);
		notifications.unsubscribe(self, ["sm.auth.change.prop.ui"]);
		notifications.unsubscribe(self, ["sm.auth.history.change"]);
	},

	notifMainLayoutChanged: function() {
		var self = this;
		var site = globals.getSiteLayout();
		var cur = site.getCurrentLayoutSettings();

		if (cur.pageParameters.forceAutoScale === true) {
			$("#s-m-auth-scale-id").addClass('active');
		} else {
			$("#s-m-auth-scale-id").removeClass('active');
		}
		if (cur.pageParameters.orientation === "landscape") {
			$("#s-m-auth-horizontal-id").addClass('active');
			$("#s-m-auth-vertical-id").removeClass('active');
		} else {
			$("#s-m-auth-horizontal-id").removeClass('active');
			$("#s-m-auth-vertical-id").addClass('active');
		}

	},

	// Called by panel if a property like color has been changed using the simple ui
	notifAuthChangePropUi: function(articlePage) {
		var self = this;
	},

	notifAuthSetSelection: function(id) {
		var self = this;
		self._currentSelectionId = id;
		self._enableRemoveBtn();
	},

	notifAuthHistoryChange: function(history) {
		var self = this;
		if (history.articles.length > 0) {
			self.$$undo.removeClass("disabled");
			self.$$undoAll.removeClass("disabled");
		} else {
			self.$$undo.addClass("disabled");
			self.$$undoAll.addClass("disabled");
		}
	},

	_enableRemoveBtn: function() {
		var self = this;
		var stateRemove = false;
		var stateRemoveCnt = false;
		var sel, data, parent;

		if (self._currentSelectionId) {
			sel = authHelpers.getSelectionById(self._currentSelectionId);
			data = sel && authHelpers.getSelectionData(sel);
			if (!data || !data.control) {
				stateRemove = false;
			} else {
				stateRemove = true;
			}
		}

		if (data && data.control && data.control) {
			if (data.control.isRoot()) { // Never allow to remove root layout node, in page and/or in vignette
				stateRemove = false;
			}

			if (data.control.isLayout()) {
				stateRemoveCnt = true;
			}
		}

		if (stateRemove !== false) {
			self.$$removeItem.removeClass("disabled");
		} else {
			self.$$removeItem.addClass("disabled");
		}
		if (stateRemoveCnt !== false) {
			self.$$removeCnt.removeClass("disabled");
		} else {
			self.$$removeCnt.addClass("disabled");
		}
	},
	_setDeviceSize: function(size, orientation) {
		var params = {
			layout: {
				pageParameters: {}
			}
		};

		if (size) {
			params.layout.pageParameters.device = size;
		}
		if (orientation) {
			params.layout.pageParameters.orientation = orientation;
		}
		authHelpers.triggerResizeInternal(params);
	},

	_onRemoveItem: function() {
		if (!this._currentSelectionId) {
			return;
		}
		var sel = authHelpers.getSelectionById(this._currentSelectionId);
		var data = sel && authHelpers.getSelectionData(sel);
		if (!data || !data.control || data.control.isRoot()) {
			return;
		}
		this._enableRemoveBtn(false);
		// Change widget tree
		data.control.parent.removeChild(data.control);
		notifications.publish(["sm.auth.change.item.prop.ui"]);
	},

	_onRemoveCnt: function() {
		if (!this._currentSelectionId) {
			return;
		}
		var sel = authHelpers.getSelectionById(this._currentSelectionId);
		var data = sel && authHelpers.getSelectionData(sel);
		if (!data || !data.control) {
			return;
		}

		data.control.removeChildren();
		notifications.publish(["sm.auth.change.item.prop.ui"]);
	},

	_onRemoveAllItems: function() {
		// Rebuild and apply new article
		var page = globals.getApplication().currentPage;
		page.rootLayout.removeChildren();
		notifications.publish(["sm.auth.change.item.prop.ui"]);
	},

	_onClose: function() {
		notifications.publish(["sm.auth.close"]);
	},

	_onSave: function(evt) {
		notifications.publish(["sm.auth.save"], $("#auth-btn-save-id").attr("data-save-type"));
	},

	_onChooseDevice: function(elmt) {
		var self = this;
		var site = globals.getSiteLayout();
		var cur = site.getCurrentLayoutSettings();

		var modal = new modalChooseDevice.Modal(cur.pageParameters.device);
		modal.show().then(function(result) {
			if (result != null) {
				self._setDeviceSize(result);
				notifications.publish(["sm.auth.display.changed"], "device");
			}
		});
	},

	_onScaleDevice: function(elmt) {
		var site = globals.getSiteLayout();
		var cur = site.getCurrentLayoutSettings();

		var params = {
			layout: {
				pageParameters: {
					forceAutoScale: cur.pageParameters.forceAutoScale === false
				}
			}
		};
		authHelpers.triggerResizeInternal(params);
		notifications.publish(["sm.auth.display.changed"], "scale");
	},

	_onOrientation: function(elmt) {
		var self = this;
		var orient = $(elmt.currentTarget).attr("data-auth-orientation");
		if (orient) {
			self._setDeviceSize(null, orient);
			notifications.publish(["sm.auth.display.changed"], "orientation");
		}
	},

	_onNewLayout: function() {
		var self = this;

		var page = globals.getApplication().currentPage;
		var data = {
			scroll: jsutils.isInstanceOf(page, "DashboardPage") ? "horizontal" : "vertical",
			template: true
		};
		var modal = new modalNewLayout.Modal(data);
		modal.show().then(function(result) {
			if (result != null) {
				notifications.publish(["sm.auth.new.layout"], result.scroll, result.template);
			}
		});
	},

	_onUndo: function() {
		notifications.publish(["sm.auth.undo"], false);
	},

	_onUndoAll: function() {
		notifications.publish(["sm.auth.undo"], true);
	}
});
