// syracuse-tablet/html/js/authoring/authoring.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var globals = require('syracuse-tablet/html/js/helpers/globals');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var notifications = require('syracuse-tablet/html/js/helpers/notifications');

var _template = '\
	<div id="authPanel" draggable="true" class="modal-dialog"> \
		<div class="modal-content"> \
			<div class="modal-header"> \
				<div class="input-group"> \
					<div id="authCode" class="form-control"></div> \
	 				<div class="input-group-btn"> \
						<button type="button" id="authPanelClose" class="btn btn-default pull-right">{{label_close}}</button> \
					</div> \
				</div> \
			</div> \
			<div class="modal-body"> \
				<ul class="nav nav-pills" role="tablist"> \
					<li class="active"><a href="#authPanelEdit" role="tab" data-toggle="tab">{{label_edit}}</a></li> \
					<li><a href="#authPanelSave" role="tab" data-toggle="tab">{{label_save}}</a></li> \
				</ul> \
				<div class="tab-content"> \
					<div id="authPanelEdit" class="tab-pane panel-body active"> \
						<textarea  id="authJSON" type="text" \
					    rows="20">{{article}}</textarea> \
					    <div class="modal-footer"> \
							<button type="button" id="authPanelApply" class="btn ">{{label_apply}}</button> \
						</div> \
					</div> \
					<div id="authPanelSave" class="tab-pane panel-body"> \
						<div class="form-group"> \
							<label>{{label_code}}</label> \
							<input class="form-control" id="authCode" /> \
						</div> \
						<div class="form-group"> \
							<label>{{label_title}}</label> \
							<input class="form-control" id="authTitle" /> \
						</div> \
						<div class="form-group"> \
							<label>{{label_description}}</label> \
							<input class="form-control" id="authDescription" /> \
						</div> \
						<label>{{label_saveas_option}}</label> \
						<div class="btn-group btn-group-justified" data-toggle="buttons"> \
	  						<div class="btn-group"> \
	  							<label class="btn btn-primary"> \
	    							<input type="radio" name="options" id="personal" checked>{{label_personal}} \
	  							</label> \
	  						</div> \
	    					<div class="btn-group"> \
	  							<label class="btn btn-primary"> \
	    							<input type="radio" name="options" id="global"> {{label_global}} \
	  							</label> \
	  						</div> \
	    					<div class="btn-group"> \
	  							<label class="btn btn-primary active"> \
	    							<input type="radio" name="options" id="factory">{{label_factory}} \
	  							</label> \
	  						</div> \
						</div> \
						<div class="modal-footer"> \
							<button type="button" id="authPanelSave" class="btn">{{label_save}}</button> \
						</div> \
					</div> \
				</div> \
			</div> \
		</div> \
	</div>';

var _authInterface;

var AuthInterface = utils.defineClass(function() {
	var self = this;
	self.$$authPanel = null;
	self.authPanelTemplate = Handlebars.compile(_template);
}, null, {
	showAuthoringPanel: function() {
		var self = this;

		self._showAuthPanel();
		self._displayAuth(globals.getApplication().currentPage);
	},

	notifPageLoaded: function(page) {
		var self = this;
		if (page && !page.isVignette) {
			self._displayAuth(page);
		}
	},

	_displayAuth: function(page) {
		var self = this;
		self.$$authCode.text(page.state.authoringName);
		self.$$authJSON.val(JSON.stringify(page.article, null, "  "));
	},

	_showAuthPanel: function() {
		var self = this;
		notifications.subscribe(self, ["sm.page.loaded"]);
		self._ensureAuthPanel();
		self.$$authPanel.show();
	},
	_hideAuthPanel: function() {
		var self = this;
		self._ensureAuthPanel();
		self.$$authPanel.hide();
		notifications.unsubscribe(self, ["sm.page.loaded"]);
	},
	_ensureAuthPanel: function() {
		var self = this;
		if (this.$$authPanel) {
			return;
		}
		var ctx = {
			label_edit: locale.text("auth.panel.label_edit"),
			label_save: locale.text("auth.panel.label_save"),
			label_close: locale.text("auth.panel.label_close"),
			label_apply: locale.text("auth.panel.label_apply"),

			label_code: locale.text("auth.panel.label_code"),
			label_title: locale.text("auth.panel.label_title"),
			label_description: locale.text("auth.panel.label_description"),
			label_saveas_option: locale.text("auth.panel.label_saveas_option"),
			label_personal: locale.text("auth.panel.label_personal"),
			label_global: locale.text("auth.panel.label_global"),
			label_factory: locale.text("auth.panel.label_factory")
		};

		self.$$authPanel = $(self.authPanelTemplate(ctx));
		$("body").append(self.$$authPanel);
		self.$$authJSON = $("#authJSON");
		self.$$authCode = $("#authCode");
		self.$$authPanelEdit = $("#authPanelEdit");
		self.$$authPanelSave = $("#authPanelSave");

		$("#authPanelSave").bind("click", self._onSave.bind(self));
		$("#authPanelClose").bind("click", self._onClose.bind(self));
		$("#authPanelApply").bind("click", self._onApply.bind(self));
	},
	_updateLayout: function() {
		var self = this;
		var page = globals.getApplication().currentPage;
		var article = self.$$authJSON.val();
		var articleParsed = JSON.parse(article);
		page.updateLayout(articleParsed).fail(function(e) {
			utils.modalError("context.sdataCommonResources error", e);
		});

	},
	_onSave: function() {
		var self = this;
	},
	_onApply: function() {
		var self = this;
		self._updateLayout();
	},
	_onClose: function() {
		var self = this;
		self._hideAuthPanel();
	}
});

exports.startDesignPage = function() {
	if (!_authInterface) {
		_authInterface = new AuthInterface();
	}

	_authInterface.showAuthoringPanel();
};
