// syracuse-tablet/html/js/authoring/authoringPanelLeft.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var globals = require('syracuse-tablet/html/js/helpers/globals');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var notifications = require('syracuse-tablet/html/js/helpers/notifications');
var authHelpers = require('syracuse-tablet/html/js/authoring/authoringHelpers');
var modal = require('syracuse-tablet/html/js/ui/modal');
var authHtml = require('syracuse-tablet/html/js/authoring/authoringHtml');
var uiutils = require('syracuse-tablet/html/js/ui/uiUtils');

exports.Panel = utils.defineClass(function() {
	var self = this;
	self.$$panel = null;
}, null, {
	show: function() {
		var self = this;

		self._ensurePanel();
		self.$$panel.show();
		self.$$splitter.show();
	},

	hide: function() {
		var self = this;
		self._ensurePanel();
		self.$$panel.hide();
		self.$$splitter.hide();
	},

	notifMainLayoutChanged: function() {
		var self = this;
	},

	_ensurePanel: function() {
		var self = this;
		if (this.$$panel) {
			return;
		}
		var ctx = {
			label_vsmall: locale.text("auth.panel.label_vsmall"),
			label_small: locale.text("auth.panel.label_small"),
			label_medium: locale.text("auth.panel.label_medium"),
			label_large: locale.text("auth.panel.label_large")
		};

		self.$$panel = $("#s-m-auth-panel-left-id");
		self.$$panel.append($(authHtml.execute("authPanelLeftStructure", ctx)));

		self._getDomObjects();
		self._initSplitter();

		$(".btn", self.$$authDeviceSize).click(function() {
			var $$this = $(this);
			var id = $("input", $$this).attr("id");
			var segs = id.split("_");
			self._setDeviceSize(segs[segs.length - 1]);
		});
		notifications.subscribe(self, ["sm.main.layout.changed"]);
	},
	_getDomObjects: function() {
		var self = this;
		self.$$authDeviceSize = $("#auth_device_size", self.$$panel);
	},
	_initSplitter: function() {
		var self = this;
		self.$$splitter = $("#s-m-auth-splitter-left-id");
		self.$$splitter.draggable({
			axis: "x",
			stop: function(e, ui) {
				var $$splitter = $(this);
				var pos = $$splitter.position();
				var width = $$splitter.width();
				var panelLeftSize = pos.left;
				uiutils.triggerResizeInternal({
					layout: {
						authPanelLeft: {
							width: panelLeftSize
						}
					}
				});
			}
		});
	},
	_setDeviceSize: function(size) {
		var self = this;
		uiutils.triggerResizeInternal({
			layout: {
				authoringParameters: {
					device: size
				}
			}
		});
	}
});
