// syracuse-x3/lib/entities/bo/boProfile.ts
"use strict";

var BoRestClient = require('../../boRestClient').BoRestClient;
var globals = require('streamline/lib/globals');

exports.entity = {
	$titleTemplate: "Business Objects profile",
	$descriptionTemplate: "Configure Business Objects profile",
	$valueTemplate: "{user} : {security}",
	$helpPage: "Administration-reference_Business-Objects-profiles",
	$properties: {
		security: {
			$title: "Security",
			$enum: [{
				$value: "secEnterprise",
				$title: "secEnterprise"
			}, {
				$value: "secLdap",
				$title: "secLdap"
			}],
			$default: "secEnterprise"
		},
		user: {
			$title: "User",
			$isMandatory: true
		},
		password: {
			$title: "Password",
			$type: "password",
			$salt: "",
			$capabilities: "confirm",
			$encrypt: true,
			$isDisabled: function(_, instance) {
				return instance.security(_) === "secLdap";
			}
		}
	},
	$relations: {
		boServer: {
			$title: "BO Server",
			$type: "boServer",
			$isMandatory: true
		}
	},
	$functions: {},
	$services: {
		test: {
			$method: "GET",
			$isMethod: true,
			$title: "Test account",
			$execute: function(_, context, instance) {
				instance.$diagnoses = instance.$diagnoses || [];
				try {
					var userProfile = globals.context.session.getUserProfile(_);
					if (!userProfile) throw new Error("Can't retrieve user profile in globals context");

					if (instance.security(_) === "secLdap" && userProfile.user(_).login(_) !== instance.user(_)) {
						throw new Error("LDAP connection can be validated only with user concerned");
					}

					var boClient = new BoRestClient(_, instance);

					var valid = true;
					if (boClient.logon(_, null, instance) == null) {
						valid = false;
						instance.$diagnoses.push({
							$severity: "error",
							$message: "Logon failed ; Get token failed"
						});
					}
					if (valid) {
						if (boClient.logoff(_) == null) {
							valid = false;
							instance.$diagnoses.push({
								$severity: "error",
								$message: "Logoff failed."
							});
						}
					}
					if (valid)
						instance.$diagnoses.push({
							$severity: "info",
							$message: instance.user(_) + "'s account is valid."
						});

				} catch (e) {
					console.error(e.stack);
					instance.$diagnoses.push({
						$severity: "error",
						$message: "" + e
					});
				}
			}
		},
	}
};
