// syracuse-x3/lib/entities/bo/boServer.ts
"use strict";
var BoRestClient = require('../../boRestClient').BoRestClient;


exports.entity = {
	$titleTemplate: "Business Objects Server",
	$descriptionTemplate: "Configure Business Objects server",
	$valueTemplate: "{name}",
	$helpPage: "Administration-reference_Business-Objects-servers",
	$properties: {
		name: {
			$title: "Name",
			$isMandatory: true,
			$linksToDetails: true,
			$isUnique: true
		},
		host: {
			$title: "Host",
			$isMandatory: true
		},
		portTomcat: {
			$title: "Tomcat port",
			$isMandatory: true,
			$default: "8080"
		},
		portRest: {
			$title: "Web services port",
			$isMandatory: true,
			$default: "6405"
		},
		adminPassword: {
			$title: "Administrator's password",
			$type: "password",
			$salt: "",
			$capabilities: "confirm",
			$isMandatory: true,
			$encrypt: true
		},
		ldapDirect: {
			$title: "Direct connection for LDAP",
			$description: "Allows to ignore totally BO profiles settings and use Syracuse LDAP user credentials when it is possible.",
			$type: "boolean",
			$default: false
		}
	},
	$relations: {},
	$functions: {
		getBoServerBaseUrl: function(_, secure) {
			return (secure ? "https://" : "http://") + [this.host(_) + ":" + this.portTomcat(_)].join("/");
		},
	},
	$services: {
		test: {
			$method: "GET",
			$isMethod: true,
			$title: "Test Administrator's account",
			$execute: function(_, context, instance) {
				instance.$diagnoses = instance.$diagnoses || [];
				try {
					var boClient = new BoRestClient(_, null, instance);

					var valid = true;
					if (boClient.logon(_, true) == null) {
						valid = false;
						instance.$diagnoses.push({
							$severity: "error",
							$message: "Logon failed ; Get token failed"
						});
					}
					if (valid) {
						if (boClient.logoff(_) == null) {
							valid = false;
							instance.$diagnoses.push({
								$severity: "error",
								$message: "Logoff failed."
							});
						}
					}
					if (valid)
						instance.$diagnoses.push({
							$severity: "info",
							$message: "Administrator's account is valid."
						});

				} catch (e) {
					console.error(e.stack);
					instance.$diagnoses.push({
						$severity: "error",
						$message: "" + e
					});
				}
			}
		},
	}
};
