// syracuse-tablet/html/js/wc/multiselection/multiSelectionMainController.js
"use strict";

var Base = require("syracuse-tablet/html/js/wc/controllers/sdataController").SdataController;

var utils = require('syracuse-tablet/html/js/wc/helpers/utils');
var sdataReq = require("syracuse-tablet/html/js/wc/sdata/sdataRequester");
var protocolHelpers = require("syracuse-tablet/html/js/wc/sdata/protocolHelpers");
var environment = require('syracuse-tablet/html/js/wc/helpers/environment');
var modules = require('syracuse-tablet/html/js/wc/common/modules');
var protoHelpers = require('syracuse-tablet/html/js/wc/sdata/protocolHelpers');

/**
 * Main multi-selection controller
 * Declared at queryPageLevel
 */
exports.MultiSelectionMainController = utils.defineClass(
	function MultiSelectionMainController(queryPage) {
		if (queryPage == null || queryPage.isFacet == null || !queryPage.isFacet("$query")) {
			throw new Error("Sdata query page expected");
		}
		Base.call(this, queryPage.pageData.dataset, null, queryPage.pageData.isWorkingCopy);
		this.page = queryPage;
		this.panel = null
		this.array = null;
	},
	Base, {

		destroy: function() {
			Base.prototype.destroy.call(this);
			this.page == null;
			this.panel = null
			this.array = null;
		},
		/**
		 * Caller queryPage
		 * Toggles the status
		 * Returns true is status enabled
		 */
		multiSelToggleStatus: function() {
			var isSmartphone = modules.get("siteLayout").getDeviceType() === "smartphone";
			if (this.panel == null) {
				var panelInfo;
				if (isSmartphone) {
					panelInfo = {
						name: "headerMultisel",
						topIndex: 0,
						selector: "header",
						$type: "application/x-panel-header-multisel-smartphone"
					}
				} else {
					panelInfo = {
						name: "footerMultisel",
						topIndex: 0,
						selector: "footer",
						$type: "application/x-panel-footer-multisel-tablet"
					}
				}
				this.panel = this.page.attachedControlsMgr.createAddtionalControl(panelInfo, {});
				this.array = this.page.getQueryArray();
			}
			this.isEnabled = this.page.attachedControlsMgr.toggleControls(isSmartphone ? "headerMultisel" : "footerMultisel", isSmartphone ? "header" : "footer");
			// Notifies header/footer
			this.panel.multiSelectionController.multiSelToggleStatus(this.isEnabled, this.array);
			return this.isEnabled;
		},
		/**
		 * Caller arrayRow/Cell on click on record
		 * Toggles the status of card/row
		 */
		multiSelToggleRecord: function($$recordElmt) {
			$$recordElmt.toggleClass("s-m-multi-selected");
			// Notifies header/footer
			this.panel.multiSelectionController.multiSelToggleRecord($$recordElmt.is(".s-m-multi-selected"));
		},
		/**
		 * Caller queryPage
		 * Applies $actionName to selected rows/cards
		 */
		multiSelTriggerAction: function($actionName) {
			alert("multiSelTriggerAction " + $actionName);
		}
	});
