// syracuse-tablet/html/js/wc/multiselection/multiSelectionMainController.js
"use strict";

var Base = require("syracuse-tablet/html/js/wc/controllers/sdataController").SdataController;

var utils = require('syracuse-tablet/html/js/wc/helpers/utils');
var jsutils = require('syracuse-tablet/html/js/wc/helpers/jsutils');
var locale = require('syracuse-tablet/html/js/wc/helpers/locale');
var sdataReq = require("syracuse-tablet/html/js/wc/sdata/sdataRequester");
var sdataUtils = require("syracuse-tablet/html/js/wc/sdata/sdataUtils");
var protocolHelpers = require("syracuse-tablet/html/js/wc/sdata/protocolHelpers");
var environment = require('syracuse-tablet/html/js/wc/helpers/environment');
var modules = require('syracuse-tablet/html/js/wc/common/modules');
var protoHelpers = require('syracuse-tablet/html/js/wc/sdata/protocolHelpers');
var sdataReq = require("syracuse-tablet/html/js/wc/sdata/sdataRequester");

/**
 * Main multi-selection controller
 * Declared at queryPageLevel
 */
exports.MultiSelectionMainController = utils.defineClass(
	function MultiSelectionMainController(queryPage) {
		if (queryPage == null || queryPage.isFacet == null || !queryPage.isFacet("$query")) {
			throw new Error("Sdata query page expected");
		}
		Base.call(this, queryPage.pageData.dataset, null, queryPage.pageData.isWorkingCopy);
		this._page = queryPage;
		// Header/Footer with multSelection button
		this._panel = null
			// $resources array
		this._array = null;
		// header/footer displayed when action are processed on selected data
		this._panelRun = null;
	},
	Base, {

		destroy: function() {
			Base.prototype.destroy.call(this);
			this._page == null;
			this._panel = null;
			this._array = null;
			this._panelRun = null;
		},
		/**
		 * Caller queryPage
		 * Toggles the status
		 * Returns true is status enabled
		 */
		multiSelToggleStatus: function() {
			var isSmartphone = modules.get("siteLayout").getDeviceType() === "smartphone";
			if (this._panel == null) {
				var panelInfo;
				if (isSmartphone) {
					panelInfo = {
						name: "headerMultisel",
						topIndex: 0,
						selector: "header",
						$type: "application/x-panel-header-multisel-smartphone"
					}
				} else {
					panelInfo = {
						name: "footerMultisel",
						topIndex: 0,
						selector: "footer",
						$type: "application/x-panel-footer-multisel-tablet"
					}
				}
				this._panel = this._page.attachedControlsMgr.createAddtionalControl(panelInfo, {});
				this._array = this._page.getQueryArray();
			}
			this.isEnabled = this._page.attachedControlsMgr.toggleControls(isSmartphone ? "headerMultisel" : "footerMultisel", isSmartphone ? "header" : "footer");
			// Notifies header/footer
			this._panel.multiSelectionController.multiSelToggleStatus(this.isEnabled);
			this._array.multiSelToggleStatus(this.isEnabled);
			if (!isSmartphone) {
				//Disable header
				this._page.attachedControlsMgr.findByName("header").disable(this.isEnabled);
			}
			return this.isEnabled;
		},
		/**
		 * Caller arrayRow/Cell on click on record
		 * Toggles the status of card/row
		 */
		multiSelToggleRecord: function($$recordElmt) {
			if (this._isProcessingData()) {
				return;
			}
			$$recordElmt.toggleClass("s-m-multi-selected");
			// Notifies header/footer
			this._panel.multiSelectionController.multiSelToggleRecord($$recordElmt.is(".s-m-multi-selected"));
		},
		/**
		 * Caller queryPage
		 * Applies $actionName to selected rows/cards
		 */
		multiSelTriggerAction: function($actionName) {
			if ($actionName === "showdiagnoses") {
				this._array.multiSelShowDiagnoses();
			} else if ($actionName === "exit") {
				this._processExit();
			} else {
				this._processStart($actionName).fail(function(error) {
					modules.get("modal").asynchError(locale.text("multiselect.processAll.fail"), error).then(function() {
						self._processEnd(_data, array);
					});
				});
			}
		},
		_isProcessingData: function() {
			return this._page.attachedControlsMgr.isVisible(this._panelRun);
		},
		/**
		 * User clicked on 'reload' icon in header/footer
		 */
		_processExit: function(array) {
			// Restore header/footer and select unselected - Remove error/Success status
			this._processTogglePanel().multiSelectionController.multiSelStopProcessing();
			// Remove diags
			this._array.multiSelRemoveDiagnoses();
		},
		/**
		 * Processes all cards/rows
		 */
		_processStart: function($actionName) {
			var self = this;
			var nbOk = 0;
			var nbKo = 0;
			return $.smResolve().then(function() {
				// Displays dedicated footer/header and hides unselected rows
				self._processTogglePanel().multiSelectionController.multiSelStartProcessing();
				self._panelRun.multiSelUpdateCounter(nbOk, nbKo);
				return {
					$actionName: $actionName,
					data: self._array.multiSelGetSelectedData(),
					link: self._page.controller.dataset.getLink($actionName)
				}
			}).then(function(info) {
				if (!info.link || !info.data) {
					return null;
				}
				var deferred = $.Deferred();
				var _exec = function(idx) {
					if (idx >= info.data.length) {
						// End
						deferred.resolve("end", info);
					} else {
						self._processOne(info, idx).then(function(status) {
							if (status === "stop") {
								deferred.resolve("stop", info);
								return;
							}
							if (status === "success") {
								nbOk++
							} else {
								nbKo++;
							}
							self._panelRun.multiSelUpdateCounter(nbOk, nbKo);
							// Next
							_exec(idx + 1);
						}).fail(function(e) {
							deferred.reject(e);
						})
					}
				};
				// First
				_exec(0);
				return deferred.promise();
			}).then(function(status, info) {
				if (status === "stop") {
					return;
				}
			})
		},
		/**
		 * Process one selected row/card
		 */
		_processOne: function(info, idx) {
			var step;
			if (info.link.$url === "$UnitTesturl") {
				step = this._procesUnitTestUrl(info, idx);
			} else {
				step = this._processUrl(info, idx)
			}
			var self = this;
			return step.then(function(success, result) {
				if (!self._isProcessingData()) {
					// Process stopped
					return "stop";
				}
				var diagsMsg = [];
				var diagsErr = [];
				if ($.isPlainObject(result)) {
					(sdataUtils.scanDiagnoses(result) || []).forEach(function(d) {
						if (d.$severity === "error") {
							diagsErr.push(d);
						} else {
							diagsMsg.push(d);
						}
					});
				}
				var result = {
					$$record: info.data[idx].$$record
				}
				success = success && diagsErr.length == 0;
				if (!success) {
					if (jsutils.isError(result)) {
						result.$diagnoses = [{
							$message: result.message,
							$stackTrace: result.stack
						}];
					} else {
						result.$diagnoses = diagsErr;
					}
				} else {
					result.$diagnoses = diagsMsg;
				}
				result.status = success ? "success" : "error";
				self._array.multiSelAddDiagnoses(result);
				return result.status;
			})
		},
		/**
		 * Execute action info.$actionName on rowData dtaSet
		 */
		_processUrl: function(info, idx) {
			var rowDataSet = info.data[idx];
			var self = this;
			var deferred = $.Deferred();
			sdataReq.triggerAction(info.$actionName, rowDataSet, this._page.isWorkingCopy).then(function(response) {
				var delta = response.data;
				if (delta) {
					// If there is a delta sent by the server, we merge it
					// This will add field level diagnoses
					// error checking will be done later independently
					return self._mergeDelta(delta).then(function() {
						return response;
					});
				}
				return response;
			}).then(function(response) {
				return protocolHelpers.computeActionResult($actionName, response);
			}).then(function(result) {
				return deferred.resolve(result.success === true, result);
			}).fail(function(e) {
				return deferred.resolve(false, jsutils.convertToDiagnoses(e));
			});
			return deferred.promise();
		},
		/**
		 * Used to test ui process.
		 * Don't send an url
		 */
		_procesUnitTestUrl: function(info, idx) {
			var deferred = $.Deferred();
			setTimeout(function() {
				if (idx % 2 == 0) {
					deferred.resolve(true);
				} else {
					deferred.resolve(false, {
						$diagnoses: [{
							$message: "test failed idx=" + idx,
							$severity: "error"
						}]
					})
				}
			}, 1000)
			return deferred.promise();
		},
		/**
		 * Create if needed and toggle the process header/footer with the stop action and 'xx successes - yy failures' message
		 */
		_processTogglePanel: function() {
			var isSmartphone = modules.get("siteLayout").getDeviceType() === "smartphone";
			if (this._panelRun == null) {
				var panelInfo;
				if (isSmartphone) {
					panelInfo = {
						name: "headerMultiselRun",
						topIndex: 0,
						selector: "header",
						$type: "application/x-panel-header-multisel-run-smartphone"
					}
				} else {
					panelInfo = {
						name: "footerMultiselRun",
						topIndex: 0,
						selector: "footer",
						$type: "application/x-panel-footer-multisel-run-tablet"
					}
				}
				this._panelRun = this._page.attachedControlsMgr.createAddtionalControl(panelInfo, {});
			}
			this._page.attachedControlsMgr.toggleControls(isSmartphone ? "headerMultiselRun" : "footerMultiselRun", isSmartphone ? "headerMultisel" : "footerMultisel");
			return this._panelRun;
		},
		/**
		 * Returns the selected dataSets
		 */
		getSelectedData: function() {
			return this._array.multiSelGetSelectedData();
		}
	});
