// syracuse-tablet/html/js/ajax/ajax.js
"use strict";
/*
 * Ajax module
 */


var ajaxInterface = require('syracuse-tablet/html/js/ajax/ajaxInterface');

var _ajax;

exports.getAjax = function() {
	if (!_ajax) {
		_ajax = new ajaxInterface.AjaxInterfaceClass();
	}
	return _ajax;
};

exports.init = function(config, locale) {
	ajaxInterface.init(config, locale);
};
exports.rejectHttpResult = function(title, result, url) {
	if (result && result.responseJSON && result.responseJSON.$diagnoses) {
		// $noConnectionError to notify caller that fail is due to network connection broke down
		if (result.responseJSON.$diagnoses.length === 0) {
			result.responseJSON.$diagnoses = [{
				$message: title,
				$severity: "error"
			}];
		}
		return $.smReject({
			$noConnectionError: result.responseJSON.$noConnectionError,
			$redirectPage: result.responseJSON.$redirectPage,
			$title: title,
			$diagnoses: result.responseJSON.$diagnoses
		});
	}
	return $.smReject({
		$title: title,
		$diagnoses: [{
			$message: "Unexpected empty data"
		}]
	});
};
