// syracuse-tablet/html/js/pages/html/modalSelectContext.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var modal = require('syracuse-tablet/html/js/ui/modal');
var globals = require('syracuse-tablet/html/js/helpers/globals');
var localStorage = require('syracuse-tablet/html/js/storage/localStorage');

var _template = '\
	<div class="modal fade" id="modal-endpoint" tabi	ndex="-1" role="dialog"> \
    <div class="modal-dialog"> \
		<div class="modal-content"> \
		    <div class="modal-header"> \
				<h4 class="modal-title" id="myModalLabel">{{title}}</h4> \
		    </div> \
		    <div class="modal-body"> \
				<div class="form-group"> \
					<select class="form-control" id="login-select-endpoint"> \
					{{#each endpoints}}\
						<option value="{{this.application}}.{{this.contract}}.{{this.dataset}}">\
						{{this.description}}\
						</option>\
					{{/each}}\
					</select>\
				</div> \
			<div class="form-group"> \
				<select class="form-control" id="login-select-role"> \
				{{#each roles}}\
					<option value="{{this.code}}">\
					{{this.description}}\
					</option>\
				{{/each}}\
				</select>\
			</div> \
			<div class="form-group"> \
				<select class="form-control" id="login-select-language"> \
				{{#each languages}}\
					<option value="{{this.code}}">\
					{{this.description}}\
					</option>\
				{{/each}}\
				</select>\
			</div> \
		    </div> \
		    <div class="modal-footer"> \
		        <button type="button" class="btn" data-action="validate">{{validate}}</button> \
    			<button type="button" class="btn btn-default" data-action="cancel">{{cancel}}</button> \
			</div> \
		</div> \
	</div> \
	</div>';

/**
 * force = true
 * 		Force display of  modal is an endpoint is still available in local storage
 */
var _selectContext = function(context, endpoints, roles, languages) {
	var deferred = $.Deferred();
	try {
		try {
			var modalHtml = Handlebars.compile(_template);
			if (endpoints.length > 0) {
				modalHtml = modalHtml({
					title: "Please, select the default endpoint",
					endpoints: endpoints,
					roles: roles,
					languages: languages,
					validate: "Validate",
					cancel: "Cancel"
				});
				modal.modal(modalHtml, function(msg, dialg) {
					if (msg === "validate") {
						var endpoint = dialg.find("#login-select-endpoint").val();
						var role = dialg.find("#login-select-role").val();
						var language = dialg.find("#login-select-language").val();

						deferred.resolve({
							endpoint: endpoint,
							role: "role",
							language: "locale"
						});
					}
				});
			} else {
				deferred.reject("No endpoint available for the user\nPlease contact administrator");
			}
		} catch (e) {
			deferred.reject(e);
		}
	} catch (e) {
		deferred.reject(e);
	} finally {
		return deferred.promise();
	}
};

exports.selectContext = _selectContext;
