// syracuse-tablet/html/js/pages/html/pageTemplates.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var globals = require('syracuse-tablet/html/js/helpers/globals');
var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("page");
var templates = require('syracuse-tablet/html/js/helpers/templating');
var ajax = require('syracuse-tablet/html/js/ajax/ajax').getAjax().ajax;

var _consts = {
	tmplPage: "pages/",
	tmplDefHeader: "pages/default/header",
	tmplRoot: "templates/pages/",
	dataFile: "/data.json",
	tmplHead: "header",
	tmplMain: "main"
};

var _defHeaderCtx = {
	title: "Sage X3 Mobile",
	home: " Home",
	about: " About",
	logout: " Logout",
	settings: " Settings",
	application: ""
};

/**
 * Reads header in templates/pageName
 * If no header reads default header in templates/default
 */
var _getTemplateHeader = function(state, ctxHeader) {
	var self = this;
	var deferred = $.Deferred();
	var _resolveError = function(e) {
		log && log(state.name + "Template header error", e);
		deferred.resolve(null);
	};
	try {
		if (ctxHeader == null) deferred.resolve(null);
		var root = _consts.tmplPage + state.name + "/";
		templates.exec(root + _consts.tmplHead, ctxHeader, {
			failNotFound: false
		}).then(function(header) {
			if (header == null && ctxHeader != null) {
				_defHeaderCtx.application = globals.getApplication().getTitle();
				templates.exec(_consts.tmplDefHeader, _defHeaderCtx).then(function(header) {
					deferred.resolve(header);
				}).fail(function(e) {
					_resolveError("Error loading default page header");
				});
			} else {
				deferred.resolve(header);
			}
		}).fail(function(e) {
			_resolveError(e);
		});
	} catch (e) {
		_resolveError(e);
	} finally {
		return deferred.promise();
	}
};

/**
 * Reads content templates in templates/pageName
 */
var _getTemplateContent = function(state, ctxContent) {
	var self = this;
	var deferred = $.Deferred();
	try {
		if (ctxContent == null) deferred.resolve(null);
		var root = _consts.tmplPage + state.name + "/";
		templates.exec(root + _consts.tmplMain, ctxContent).then(function(content) {
			deferred.resolve(content);
		}).fail(function(e) {
			_resolveError(deferred, state.name, e);
		});
	} catch (e) {
		_resolveError(deferred, state.name, e);
	} finally {
		return deferred.promise();
	}
};


/**
 * Mockup - Load a data.json file in templates/pageName
 */
var _getData = function(state) {
	var deferred = $.Deferred();
	var url = globals.baseLocation().htmlRoot + state.name + _consts.dataFile;
	if (state.detailid) {
		// If the link has a data-detailid attribute we expect a data.detaild.json file
		url = url.replace(".json", "." + state.detailid + ".json");
	}
	ajax("GET", url).then(function(result) {
		deferred.resolve(result.status === 200 ? result.responseJSON : null);
	}).fail(function(e) {
		deferred.reject(e);
	});
	return deferred.promise();
};

var _resolveError = function(deferred, pageName, e) {
	var content = '<div class="alert alert-danger" role="alert"> Error loading page  "' + pageName + '"</div>';
	if (e) {
		content += "<p>Error</p><pre>" + JSON.stringify(utils.toDiagnose(e, null, true), null, 2) + "</pre>";
	}
	if (deferred) {
		deferred.resolve(content);
	}
};

exports.getTemplateContent = _getTemplateContent;
exports.getTemplateHeader = _getTemplateHeader;
exports.getData = _getData;
exports.resolveError = _resolveError;
