// syracuse-tablet/html/js/pages/html/pageHtml.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("page");
var uiutils = require('syracuse-tablet/html/js/ui/uiUtils');
var Base = require('syracuse-tablet/html/js/pages/page').Page;
var pageTemplates = require('syracuse-tablet/html/js/pages/pageTemplates');

/**
 * Temporarily HTML page
 * 		All pages should be generated based on $article and $prototype
 */
var _Page = utils.defineClass(
	function($parent, state) {
		var self = this;
		Base.call(self, $parent, state);
	}, Base, {
		/**
		 * Header comes from html
		 */
		_createPageHeader: function() {
			return null;
		},
		/**
		 * No article an proto
		 */
		loadStructure: function() {
			return $.Deferred(function() {
				this.resolve();
			}).promise();
		},

		buildHtml: function(contentCtx, headerCtx) {
			var deferred = $.Deferred();
			var self = this;
			try {
				$.when(self.buildHtmlContent(contentCtx), self.buildHtmlHeader(headerCtx)).then(function(content, header) {
					deferred.resolve({
						content: content,
						header: header
					});
				}, function(e) {
					deferred.reject(e);
				});
			} catch (e) {
				deferred.reject(e);
			} finally {
				return deferred.promise();
			}
		},

		buildHtmlContent: function(ctxContent) {
			var self = this;
			return pageTemplates.getTemplateContent(self.state, ctxContent || {});
		},

		buildHtmlHeader: function(ctxHeader) {
			var self = this;
			return pageTemplates.getTemplateHeader(self.state, ctxHeader || {});
		}

	});

exports.Page = _Page;
