// syracuse-tablet/test/js/main.js
"use strict";
require('syracuse-tablet/html/js/helpers/nativePrototypes');
var utils = require('syracuse-tablet/html/js/helpers/utils');
var modal = require('syracuse-tablet/html/js/ui/modal');
var storage = require('syracuse-tablet/html/js/storage/storage');
var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("storage");

var _htmlRooturl = (function() {
	var url = window.location.href.split('/');
	url.splice(url.length - 2, 2);
	url.push('html');
	return url.join('/');
})();

var _runModules = [
	require('../../test/js/storage/testStorage'),
	require('../../test/js/testAjax'),
	require('../../test/js/testGeoLoc'),
	require('../../test/js/sdata/testCache'),
	require('../../test/js/sdata/where/testWhere'),
	require('../../test/js/sdata/testStateless'),
	require('../../test/js/sdata/testDispatch'),
	require('../../test/js/testUtils')
];

exports.main = function() {

	var _globalCtx = new(function() {
		var _self = this;
		var _data = null;
		_self.$rootHeader = $("#testsheader");
		_self.$rootContent = $("#testscontent");
		_self.$dropDownMenu = $("#run_dropdown");
		_self.save = function() {
			storage.localStorage.setItem("smTestContext", JSON.stringify(_data));
		};
		_self.getLogin = function() {
			return _data.login;
		};
		_self.setLogin = function(user, pwd) {
			_data.login = {
				user: user,
				password: pwd
			};
			_self.save();
		};
		_self.setEndpoint = function(type, value) {
			type = type || "";
			value = value || "";
			if (type.length == 0 || value.length == 0) return;
			_data.endpoints[type] = value;
			_self.save();
		};
		_self.getEndpoint = function(type) {
			return _data.endpoints[type];
		};
		_data = storage.localStorage.getItem("smTestContext");
		if (_data != null && typeof _data == "string") {
			try {
				_data = JSON.parse(_data);
			} catch (e) {
				log && log("Error parsing global context");
				_data = null;
			}
		} else {
			_data = null;
		}
		if (_data == null) {
			_data = {
				login: {
					user: "guest",
					password: "guest"
				},
				endpoints: {
					DVLP: "SUPERV"
				}
			};
		}
		window.globalCtx = _self;
		_self.$rootHeader.find("#nav-login-info").html("Login: " + _data.login.user + "/" + _data.login.password + " - Enpoint: " + _self.getEndpoint("DVLP"));
		log && log("global context initialized");
	});

	var _doSettings = function() {
		modal.modal("settings", {}, function($html, act, params) {
			var pref = "login-";
			var login = _globalCtx.getLogin();
			var user = $html.find("#" + pref + "user"),
				pwd = $html.find("#" + pref + "password"),
				dvlp = $html.find("#" + pref + "endpoint-dvlp");
			if (act === "shown.bs.modal") {
				user.val(login.user);
				pwd.val(login.password);
				dvlp.val(_globalCtx.getEndpoint("DVLP"));
			} else if (act === "save") {
				_globalCtx.setLogin(user.val(), pwd.val());
				_globalCtx.setEndpoint("DVLP", dvlp.val());
			}

		});
	};
	var _bindEvents = function() {
		$(document).delegate(_globalCtx.$rootContent, "click", function(evt) {
			var target = $(evt.target);
			var elmt = target.closest("[data-action]");
			var options = {};
			if (elmt.length > 0) {
				var attr = elmt.attr("data-action");
				var params = elmt.attr("data-param");
				evt.preventDefault();
				evt.stopPropagation();
				if (attr === "settings") {
					_doSettings();
				} else if (attr === "run") {
					_run(params);
				}
			}
		});
	};
	var _runCpt = 0;
	var _run = function(modulename) {
		var $runBtn = _globalCtx.$rootHeader.find("#testsrun");
		$runBtn.button('loading');
		_globalCtx.$rootContent.unbind();
		_globalCtx.$rootContent.empty();
		window.setTimeout(function() {
			_runCpt++;
			try {
				var modules = $.extend([], _runModules);
				var deferreds = [];
				modules.forEach(function(m) {
					if (modulename === 'all' || m.modulename === modulename) {
						deferreds.push(m.test());
					}
				});
				$.when.apply(null, deferreds).then(function() {
					// Called if all modules succeed
					$runBtn.button('reset');
					log && log("RUNTEST STOP - OK");
				}, function(e, modulename) {
					// Called if one module fails
					$runBtn.button('reset');
					modulename = modulename || "noName";
					log && log("RUNTEST STOP - EXCEPTION - Module " + modulename);
					log && log(e);
					modal.error("Run test exception - Module " + modulename, e);
				});
			} catch (e) {
				$runBtn.button('reset');
				modal.error("Run test error", e);
			}
		}, _runCpt == 0 ? 0 : 300);
	};

	var _init = function() {
		try {
			modal.init(_htmlRooturl);

			_runModules.forEach(function(m) {
				if (m && m.modulename) {
					$('<li><a data-action="run"  data-param="' + m.modulename + '" href="#">Test ' + m.modulename + '</a></li>').appendTo(_globalCtx.$dropDownMenu);
				}
			});
			// To avoid bootstrap transitions (collapse)
			$.support.transition = false;
			_bindEvents();
			var param = /runTest=(.*)/.exec(document.location.href);
			if (param && param[1]) {
				_run(param[1]);
			}
		} catch (e) {
			console.log(e);
			alert("Runtest error\n" + e.message);
		}
	};
	$(document).ready(function() {
		_init();
	});

};
