// syracuse-tablet/test/js/testUtil.js
"use strict";

var _testId = 0;

exports.addTestset = function(title) {
	var id = _getId();
	var html = _createHtml("#testset", {
		id: id,
		title: title
	});

	var $html = $(html);
	$html.appendTo("#testsets");

	return {
		id: id,
		title: title,
		$root: $html
	};
};

exports.addTestcase = function(testset, title) {
	var id = _getId();
	var html = _createHtml("#testcase", {
		id: id,
		title: title
	});

	var $html = $(html);
	$html.appendTo(testset.$root);

	return {
		id: id,
		title: title,
		$testset: testset.$root,
		$root: $html
	};
};

exports.addResult = function(testcase, title, status, error) {
	var html = _createHtml("#teststatus", {
		icon: status ? "ok" : "remove",
		title: title,
		error: status ? "&nbsp;" : (error ? ("<p>" + error + "</p>") : "&nbsp;")
	});
	if (!status) {
		var $status = testcase.$root.children(".glyphicon-ok");
		$status.removeClass("glyphicon-ok");
		$status.addClass("glyphicon-remove");

		$status = testcase.$testset.children(".panel-heading").find(".glyphicon-ok");
		$status.removeClass("glyphicon-ok");
		$status.addClass("glyphicon-remove");

		$status = testcase.$testset;
		$status.removeClass("panel-success");
		$status.addClass("panel-danger");

	}

	var $html = $(html);
	$html.appendTo(testcase.$root.find("ul"));

	return {
		$root: $html
	};
};

function _getId() {
	_testId = _testId + 1;
	return _testId;
}

function _createHtml(selector, data) {
	return $(selector).html().replace(/\{([\w-]+)\}/g, function(match, p1) {
		return data[p1] || match;
	});
}
