// syracuse-ui/lib/page/aside/breadcrumbItem.js
"use strict";

function BreadcrumbItem(page, index, slot) {
	syra_item.initialize(page, this);
	this.page = page;
	this.slot = slot;
	this.index = index;
	this.$bind = "page_breadcrumb";
	this.$item = {};
}

BreadcrumbItem.prototype.add = function(title, last, readonly) {
	var separator = syra_dom.span();

	if (readonly) {
		this.item = syra_dom.span(this.page.$skin + "-breadcrumb-item");
	} else {
		this.item = syra_dom.anchor(this.page.$skin + "-breadcrumb-item");
		this.item.syraItem = this.id;
		this.item.syraOnClick = "click";
	}

	this.item.textContent = title;

	if (!last) {
		separator.className = "s-page-breadcrumb-item s-fonticon-btn";
		separator.innerHTML = syra_context.isRTL ? syra_icon.font["filter_lt"] : syra_icon.font["filter_gt"];
		this.item.appendChild(separator);
	}


	this.slot.appendChild(this.item);
};

BreadcrumbItem.prototype.click = function(event) {
	syra_menus.click.fire({
		menu: this,
		event: event
	});
};

exports.add = function(page, index, slot, title, last) {
	return new BreadcrumbItem(page, index, slot).add(title, last, page.overPanel && page.overPanel.isModal);
};
