// syracuse-ui/lib/page/aside/megaBar.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var BarResizerDDAgent = require("./barResizerDDAgent").BarResizerDDAgent;
//var bootStrap = require('syracuse-ui/deps/bootstrap/js/bootstrap');

function _appplyPreferences(bar) {
	var state;
	if (syra_site.$preferences) {
		state = syra_site.$preferences[bar.preferenceKey || bar.userPreferenceKey || "none"];
	}
	state = state || {
		$isCollapsed: false
	};
	if (bar.preferenceKey) {
		syra_site.getPreferences(bar.preferenceKey, state, bar);
	} else {
		if (bar.userPreferenceKey) {
			state = bar.page.userPreferences.getForKey(bar.userPreferenceKey) || state;
		}
		syra_site.applyValues(bar, state);
	}
	bar.$isFloating = bar.options.$isFloating || syra_site.isTabletDevice;
}

function MegaBar() {}

exports.MegaBar = helpers.defineClass(MegaBar, null, {
	load: function(page) {
		//syra_site.loadStyleSheet("bootstrap/css/bootstrap.css");
		this.articleParent = page;
		if (!this.id) {
			this.id = page.id + "-" + (++page._childItemOffset);
		}
		syra_store.add(this);
		this.closeWidth = 100;
		this.minWidth = '100%'; //this.minWidth || 200;
		//this.slot.height = "50px";
		this.page = page;
		_appplyPreferences(this);

		this.options.resizeDirection = this.options.resizeDirection || "left";
		//this.slot.className += " s-bar-slot s-bar-slot-" + this.options.resizeDirection;
		if (this.$width != undefined && parseInt(this.$width) <= this.closeWidth) {
			delete this.$width;
		}
		if (this.options.$isResizerDisabled) {
			this.options.isAutoModeDisabled = true;
		}
		if (!this.options.$isResizerDisabled || this.options.$isCollapsable) {
			this.resizePicker = document.createElement("a");
			this.resizePicker.className = this.options.$skin + "-resizer s-bar-resizer-" + this.options.resizeDirection;
			this.openButton = syra_menus.button.add({
				parent: this,
				slot: this.resizePicker,
				text: syra_local.bar_collapse,
				css: this.options.$skin + "-opener s-bar-opener",
				iconOnly: true,
				shortCutTip: ((this.options.resizeDirection == "left") ? syra_shortCuts.tip.hideLeftBar : syra_shortCuts.tip.hideRightBar) + ", " + syra_shortCuts.tip.hideFull,
				fontIcon: this.options.resizeDirection + "_arrow",
				btnclick: function() {
					this.parent.collapse();
				}

			});
			if (!this.options.$isResizerDisabled) {
				this.resizePicker.syraDragSpot = this.id;
			}
			this.slot.appendChild(this.resizePicker);
			syra_dd.addToBarResizers(this, true);
		}

		this.slot.appendChild(this.body);
		syra_dom.setZIndex(this.slot, true);
		this.page.setArticleId(this.slot);
		this.page.bars.push(this);
		this.ensureState();
	},
	isDraggable: function(target, event) {
		if (target == this.resizePicker) {
			if (event.target != this.openButton.link) {
				syra_dd.dropableItem = {
					megaBar: this,
					resizeDirection: this.options.resizeDirection
				};
				syra_dd.start(this, BarResizerDDAgent, this.slot);
				event.syraRetValue = false;
				return;
			}
		}
		return null;
	},
	resizeSplitter: function(newWidth) {
		if (!this.designing) {
			if (!this.options.isAutoModeDisabled) {
				if (this.isSmall != syra_site.getSize().isSmall) {
					this.isSmall = syra_site.getSize().isSmall;
					this.$isCollapsed = this.isSmall;
				}
			}!newWidth && this.ensureState();
			if (this.slot.parentNode) {
				if (newWidth !== undefined) {
					this.slot.style.width = newWidth + "px";
				}
				var direction = this.options.resizeDirection;
				if (this.$isCollapsed || !this.$isFloating) {
					if (syra_site.isDocumentRTL) {
						direction = direction == "left" ? "right" : "left";
					}
					if (this.slot.parentNode) {
						//this.slot.parentNode.style["padding" + direction.substr(0, 1).toUpperCase() + direction.substr(1)] = (newWidth || this.slot.clientWidth) + "px";
					}
				}
			}!newWidth && this.resizeBar();
		}
	},

	toggleBar: function(show) {
		syra_dom.hide(this.slot, this.isHidden = !show);
	},
	ensureState: function(startResize) {
		if (!this.isHidden) {
			if (startResize && this.$isCollapsed) {
				this.curCollapsedState = this.$isCollapsed = false;
				delete this.$width;
				this.slot.style.width = this.slot.clientWidth + "px";
				syra_dom.hide(this.body, false);
			} else {
				if (this.curCollapsedState != this.$isCollapsed) {
					if (this.$isCollapsed) {
						syra_dom.hide(this.body, true);
						this.slot.style.width = "";
					} else {
						this.$isCollapsed = false;
						this.slot.style.width = '100%'; //Math.max(this.$width || 0, this.minWidth) + "px";
						this.slot.style.height = "44px";
						syra_dom.hide(this.body, false);
					}
					this.curCollapsedState = this.$isCollapsed;
				}
			}
			syra_dom.toggleClass(this.slot, "s-close", this.$isCollapsed);
			if (this.resizePicker) {
				var icon = this.options.resizeDirection;
				if (this.$isCollapsed) {
					icon = icon == "left" ? "right" : "left";
				}
				syra_menus.button.setText(this.openButton, this.$isCollapsed ? syra_local.bar_expand : syra_local.bar_collapse, icon + "_arrow");
			}
		}
	},
	resizeBar: function(resize) {

		if (this.slot) {

			if (!this.moreBtn) {

				this.moreBtnSlot = document.createElement("div");
				this.moreBtnSlot.className = "s-mega-menu-links";
				this.field.body.parentNode.appendChild(this.moreBtnSlot);

				this.moreBtn = syra_menus.button.add({
					parent: syra_site,
					slot: this.moreBtnSlot,
					text: syra_local.megaMenuMore,
					title: syra_local.megaMenuMoreTooltip,
					css: "s-mega-more-btn",
					btnclick: function() {}
				});
			}

			var maxW = this.slot.clientWidth;
			// have we overflowed?
			if (this.field.body.scrollWidth > this.slot.clientWidth) {
				var nodes = this.field.body.childNodes;
				var usableWidth = 0;
				for (var ii = nodes.length - 1; ii > -1; ii--) {
					if ((nodes[ii].getBoundingClientRect().right + this.moreBtn.link.clientWidth) < maxW) {
						break;
					} else {
						this.field.body.removeChild(nodes[ii]);
					}
				}
			}

			// var buttonSlotWidth = Math.max(0, Math.max(0, maxW - this.slot.getBoundingClientRect().left) - this.moreBtn.link.clientWidth) + "px";
			// if (this.moreBtnSlot.clientWidth <= (this.moreBtnSlot.scrollWidth - 2)) {

			// 	syra_menus.button.visibility(this.moreBtn, false);
			// 	this.body.style.height = "44px";
			// }
		}
		return false;
	},

	collapse: function() {
		delete this.onBeforeMaximized;
		this.$isCollapsed = !this.$isCollapsed;
		this.savePreferences();
		this.ensureState();
		this.page.resizeArticle(true);
	},
	clearPreferences: function() {
		var key = this.preferenceKey || this.userPreferenceKey;
		if (key) {
			syra_site.setPreferences(key);
			delete this.$width;
			delete this.$isCollapsed;
			this.ensureState();
			_appplyPreferences(this);
		}
	},
	savePreferences: function(key) {
		if (this.preferenceKey) {
			syra_site.setPreferences(this.preferenceKey, this, ["$width", "$isCollapsed"]);
		} else {
			if (this.userPreferenceKey) {
				var pref = this.page.userPreferences.getForKey(this.userPreferenceKey, {});
				pref.$width = this.$width;
				pref.$isCollapsed = this.$isCollapsed;
				this.page.savePageDesign(true);
			}
		}
	},
	dispose: function() {
		var bars = this.page && this.page.bars;
		if (bars) {
			var index = bars.indexOf(this);
			if (index >= 0) {
				bars.splice(index, 1);
			}
		}
		syra_store.remove(this);
		this.resizePicker && syra_dd.addToBarResizers(this, false);
		this.$isFloating && syra_dom.removeChild(this.slot);
		syra_site.disposeObject(this);
	}



	// resize: function(slot) {
	// 	// var maxW = syra_site.headerBottom.clientWidth;
	// 	// var x = slot.getBoundingClientRect();
	// 	//_manager.buttonsSlot.style.width = Math.max(0, Math.max(0, maxW - _manager.buttonsSlot.getBoundingClientRect().left) - _manager.moreBtn.link.clientWidth) + "px";
	// 	//if (_manager.buttonsSlot.clientWidth <= (_manager.buttonsSlot.scrollWidth - 2)) {
	// 	//}

	// }
});
