// syracuse-ui/lib/desktop/site/authoring/authoringView.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Article = require("syracuse-ui/lib/common/article/article").Article;
var ItemsTree = require('./itemsTree').ItemsTree;
var DesignPalette = require("./designPalette").DesignPalette;

var _localize = {
    title: "Authoring mode",
    close: "Close",
    fullViewMode: "Full view"
};

function AuthoringView(){
}

exports.AuthoringView = helpers.defineClass(AuthoringView, Article, {
    drawBox: function(){
        var self = this;
        self.$$container.empty();
        self.tree = new ItemsTree();
        self.tree.load(self, self.$$container);
        
        var $$header = $("<div id='s-author-header'/>").appendTo(self.$$container);
        var $$title = $("<div id='s-author-header-title'/>").text(_localize.title).appendTo($$header);
        $$title.bind("click", function(){
            self.$isCancelAuthoringDisabled = true;
            return false;
        });
        var $$body = $("<div id='s-author-header-body'/>").appendTo($$header);
        var $$actions = $("<div class='s-author-header-actions'/>");
        document.itemFactory.load($$actions, {
            $bind: "$save",
            $category: "link",
            $skin: "s-author-save"
        }, self);
        self.headerMenus = document.itemFactory.load($$actions, {
            $category: "links",
            $skin: "s-author-header-links",
            $excludeBind: ["$details", "$query", "$print"]
        }, self);
        $$actions.appendTo($$body);
        var close = document.itemFactory.load($$body, {
            $bind: "$close",
            $category: "link",
            $noText: true,
            $skin: "s-author-close"
        }, self);
        close.onMenuClick = function(){
            if (document.authoringSite.$isPageUpdated && !self.$isCancelAuthoringDisabled) {
                var $sourceItem = document.authoringSite.$sourceTargetPageItem;
                document.site.showMessage({
                    $title: "Page layout has been updated",
                    $message: "Save Update before quit",
                    $type: "question",
                    $buttons: "yesnocancel",
                    callback: function(response){
                        if (response.$id == "yes") {
                            self.menusFacade.items.$save[0].click();
                        }
                        else {
                            if (response.$id == "no") {
								document.authoringSite.toggle(false);
								document.controller.reloadMainPage();
                            }
                        }
                    }
                });
            }
            else {
                document.authoringSite.toggle(false);
            }
            return false;
        };
        close.setMenu({
            $title: _localize.close
        }, null);
        
        var $$body = $("<div id='s-author-body'/>").appendTo(self.$$container);
        self.$$path = $("<nav class='s-author-path'/>").appendTo($$body);
        self.$$panels = $("<div class='s-author-panels'/>").appendTo($$body);
        
        var palette;
        if (this.tree && this.tree.designedNode && (palette = this.tree.designedNode.palette)) {
            palette.$$container = this.$$panels;
            palette.$$item[0].style.height = '';
            palette.$$item.appendTo(palette.$$container);
        }
        document.site.resize();
        
        self.tree.loadArticle(document.authoringSite.targetPage);
    },
    onNotifyRecordChange: function(value, binding){
        var data;
        if (binding == "$designPalette") {
            document.authoringSite.$isPageUpdated = true;
            data = null;
        }
        else {
            data = {};
            /*if (!this.dataset.application) {
             var httpQuery = document.controller.parseUrl(document.authoringSite.targetPage.$prototype.$representationUrl);
             data.application = httpQuery.$urlParts.endpointParts.application;
             data.contract = httpQuery.$urlParts.endpointParts.contract;
             data.representation = httpQuery.$urlParts.representationRoot;
             data.facet = httpQuery.$urlParts.$facet;
             }*/
            debugger;
            data.content = {
                $article: document.authoringSite.targetPage.$item
            };
            data[binding] = value;
        }
        return data;
    },
    createPalette: function(designedNode){
        var palette = designedNode.palette = new DesignPalette();
        palette.$$container = this.$$panels;
        palette.$prototype = {};
        palette.designedNode = designedNode;
        document.itemFactory.initializeItem(palette, {
            $bind: "$designPalette"
        }, this);
        palette.loadBox();
        return palette;
    },
    dispose: function(){
        if (this.$$path) {
            this.$$path.undelegate();
        }
        delete this.page;
        Article.prototype.dispose.call(this);
    }
});
