// syracuse-ui/lib/desktop/site/authoring/authoring.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');


var AuthoringView = require("./authoringView").AuthoringView;

var _pounds = {
    "article": 4,
    "section": 3,
    "block": 2,
    "field": 1
};

function AuthoringSite(){
}

exports.AuthoringSite = helpers.defineClass(AuthoringSite, null, {
    getPound: function($authoringLevel){
        return _pounds[$authoringLevel || "field"];
    },
    toggle: function(show){
        var self = this;
        document.site._$$header.toggle(!show);
        self.$$container.toggle(show);
        if (show) {
            self.isAuthoringOn = true;
            self.article = new AuthoringView();
            self.article.$$container = self.$$container;
            self.$sourceTargetPageItem = (self.targetPage = document.site.mainPage).$item;
			self.$isPageUpdated = false;
            document.controller.loadWorkingCopy({
                menu: {
                    $url: self.targetPage.$authorUrl
                },
                article: self.article,
                callback: function(){
                    document.site.resize();
                }
            });
        }
        else {
            delete self.isAuthoringOn;
            document.controller.disposeObject(self.article);
            self.$$container.empty();
            self.article = null;
            document.site.resize();
        }
    },
    createPalette: function(designedNode){
        return this.article.createPalette(designedNode);
    },
    dispose: function(){
        delete this.$sourceTargetPageItem;
        delete this.targetPage;
    }
});
