// syracuse-ui/lib/desktop/site/authoring/itemPalette.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Article = require("syracuse-ui/lib/common/article/article").Article;
var layoutSettings = require('syracuse-ui/lib/common/article/layouts/authoring/layoutSettings');
var _localize = {
    allFields: "All fields",
    allSections: "All sections",
    layoutStacked: "stacked",
    layoutInTab: "in tab",
    layoutSide: "side by side",
    addSection: "add a section",
    lightMode: "show to light mode",
    fullMode: "show to full mode",
    isSeparatorsVisible: "show separators",
    widthContent: "content",
    widthSmall: "small",
    widthMedium: "medium",
    widthLarge: "large",
    widthMaximum: "maximum",
    isHidden: "hide me",
    labelLeftAlignment: "at left",
    labelTopAlignment: "at top",
    textLabelAlignment: "text alignment",
    textLabelLeftAligned: "left",
    textLabelRightAligned: "right",
    layoutColumn: "1 column",
    layoutColumns: "columns",
    layoutComposite: "composite",
    isBoxCollapsable: "collapsible",
    isMaximizable: "maximizable",
    isTitleHidden: "hide title",
    isTopLabelAlignment: "top aligned",
    isLabelHidden: "hide label",
    allBlocks: "All blocks",
    addBlock: "add a block",
    fieldsPosition: "position",
    addField: "add a field",
    fieldSize: "Size",
    fieldLabelPosition: "Labels position",
    layout: "Layout",
    appearance: "Appearance",
    structure: "Structure"
};




function _formatDefaultItem($bind){
    return {
        $bind: $bind,
        $skin: "s-author-field",
        $isEditMode: true,
        $isLeftSpaceHidden: true
    };
}

function _formatLinks($protoAddItem){
    return $protoAddItem ? {
        $switchMode: {
            $title: _localize.lightMode
        },
        $addItem: {
            $title: $protoAddItem
        }
    } : {
        $layoutType: "columns",
        $layoutSubType: "50,50",
        $items: [{
            $layoutType: "stack",
            $items: [{
                $bind: "$switchMode",
                $skin: "s-author-designer-link",
                $css: "s-author-designer-link-switch-light",
                $category: "link"
            }]
        }, {
            $layoutType: "stack",
            $items: [{
                $bind: "$addItem",
                $skin: "s-author-designer-link",
                $category: "link"
            }]
        }]
    };
}

function _formatBoxLayoutStack(){
    return {
        $layoutType: "stack",
        $items: [{
            $layoutType: "columns",
            $layoutSubType: "33,33,33",
            $items: [{
                $layoutType: "stack",
                $items: [_formatLayout(false)]
            }, _formatBoxAppearance(), {
                $layoutType: "stack",
                $items: [{
                    $category: "section",
                    $title: _localize.structure,
                    $layout: {
                        $items: ["$isBoxCollapsable", "$isMaximizable"].map(function($bind){
                            return _formatDefaultItem($bind);
                        })
                    }
                }]
            }]
        }, _formatLinks()]
    };
}

function _formatLayout(isProto){
    if (isProto) {
        var $enums = layoutSettings.$columns.map(function($width){
            var $parts = $width.split(",");
            return {
                $value: $parts.join("-"),
                $title: $parts.length + " " + _localize.layoutColumns // + " ( " + $parts.join("% ") + ")"
            };
        });
        $enums.splice(0, 0, {
            $value: "100",
            $title: _localize.layoutColumn
        });
        Object.keys(layoutSettings.$composite).forEach(function($key){
            $enums.push({
                $value: $key,
                $title: layoutSettings.$composite[$key].$title || _localize.layoutComposite
            });
        });
        return {
            $type: "application/x-choice",
            $value: {
                $type: "application/x-string",
                $constraints: {
                    $enum: $enums
                }
            }
        };
    }
    return {
        $category: "section",
        $title: _localize.layout,
        $layout: {
            $items: [{
                $bind: "$boxLayoutType",
                $iconMode: "iconText",
                $iconPath: "authoring/s-layout-",
                $iconSize: {
                    $width: "32px",
                    $height: "32px"
                },
                $choiceLayout: "3",
                $isEditMode: true,
                $skin: "s-author-field",
                $isLeftSpaceHidden: true
            }]
        }
    };
}

function _formatFiewWidth(isProto){
    return isProto ? {
        $type: "application/x-choice",
        $value: {
            $type: "application/x-string",
            $constraints: {
                $enum: [{
                    $value: "content",
                    $title: _localize.widthContent
                }, {
                    $value: "small",
                    $title: _localize.widthSmall
                }, {
                    $value: "medium",
                    $title: _localize.widthMedium
                }, {
                    $value: "large",
                    $title: _localize.widthLarge
                }, {
                    $value: "fit",
                    $title: _localize.widthMaximum
                }]
            }
        }
    } : {
        $category: "block",
        $title: _localize.fieldSize,
        $layout: {
            $items: [{
                $bind: "$field$width",
                $skin: "s-author-field",
                $isEditMode: true,
                $isTitleHidden: true,
                $labelWidth: "auto",
                $format: "$combo"
            }]
        }
    };
}


function _formatFieldLabelAlign(isProto){
    return isProto ? {
        $type: "application/x-choice",
        $value: {
            $type: "application/x-string",
            $constraints: {
                $enum: [{
                    $value: false,
                    $title: _localize.labelLeftAlignment
                }, {
                    $value: true,
                    $title: _localize.labelTopAlignment
                }]
            }
        }
    
    } : _formatDefaultItem("$field$isTopLabelAlignment");
}

function _formatFieldTextLabelAlign(isProto){
    return isProto ? {
        $type: "application/x-choice",
        $title: _localize.textLabelAlignment,
        $value: {
            $type: "application/x-string",
            $constraints: {
                $enum: [{
                    $value: false,
                    $title: _localize.textLabelLeftAligned
                }, {
                    $value: true,
                    $title: _localize.textLabelRightAligned
                }]
            }
        }
    } : {
        $bind: "$field$isRightTextLabelAlignment",
        $skin: "s-author-field",
        $isEditMode: true,
        $labelWidth: "auto"
    };
}

function _formatFieldsLabelPositionBlock(){
    return {
        $category: "block",
        $title: _localize.fieldLabelPosition,
        $layout: {
            $items: [_formatFieldLabelAlign(false), _formatFieldTextLabelAlign(false)]
        }
    };
}

function _formatBoxAppearance(){
    return {
        $layoutType: "stack",
        $items: [{
            $category: "section",
            $title: _localize.appearance,
            $layout: {
                $items: ["$isHidden", "$isSeparatorsVisible", "$isTitleHidden"].map(function($bind){
                    return _formatDefaultItem($bind);
                })
            }
        }]
    };
}

function _formatChildLayout(isProto, $category){
    if (isProto) {
        var $enum = [{
            $value: "stack",
            $title: _localize.layoutStacked
        }];
        if ($category != "field") {
            $enum.push({
                $value: "tabs",
                $title: _localize.layoutInTab
            });
        }
        $enum.push({
            $value: "side",
            $title: _localize.layoutSide
        });
        return {
            $type: "application/x-choice",
            $value: {
                $type: "application/x-string",
                $constraints: {
                    $enum: $enum
                }
            }
        };
    }
    else {
        return {
            $category: $category == "field" ? "block" : "section",
            $bind: "$field",
            $title: {
                section: _localize.allSections,
                block: _localize.allBlocks,
                field: _localize.fieldsPosition
            }[$category],
            $layout: {
                $items: [{
                    $bind: "$allChildLayout",
                    $skin: "s-author-field",
                    $format: "$radios",
                    $iconMode: "iconText",
                    $choiceLayout: "1",
                    $iconPath: "authoring/s-layout-",
                    $iconSize: {
                        $width: "32px",
                        $height: "32px"
                    },
                    $isEditMode: true,
                    $isTitleHidden: true,
                    $labelWidth: "auto"
                }]
            }
        };
    }
}

function _addNewBox(parentBox, $category){
    var newItem = document.itemFactory.create({
        $category: $category,
        $title: _localize["new" + $category.slice(0, 1).toUpperCase() + $category.slice(1)],
        $layout: {
            $items: []
        }
    }, parentBox);
    parentBox.layoutContent.appendNewItem({
        newItem: newItem,
        load: true
    });
    newItem.authoringNode.tree.showTargetItem(newItem.authoringNode);
}

function _notifyChildrenFields(children, value, binding){
    children.forEach(function(child){
        if (child.$authoringLevel == "field") {
            _notifyFieldItem(child.item, value, binding);
        }
        else {
            if (child.children) {
                _notifyChildrenFields(child.children, value, binding);
            }
        }
    });
}

function _notifyFieldItem(item, value, binding){
    var metaData = {}
    metaData[binding.slice("$field".length)] = value;
    item.applyDesignMetaData(metaData);
    if (item.authoringNode && metaData.$isHidden !== undefined) {
        item.authoringNode.applyHiddenState(metaData);
    }
}

function ItemPalette(){

}


helpers.defineClass(ItemPalette, Article, {
    getPage: function(){
        return this;
    },
    onNotifyRecordChange: function(value, binding){
        var self = this;
        var metaData = {};
        metaData[binding] = value;
        if (binding.indexOf("$field$") == 0) {
            if (self.designedNode.$authoringLevel == "field") {
                _notifyFieldItem(self.designedNode.item, value, binding);
            }
            else {
                _notifyChildrenFields(self.designedNode.children, value, binding);
                self.designedNode.item.applyDesignMetaData(metaData);
            }
        }
        else {
            if (metaData.$boxLayoutType !== undefined) {
                metaData.$layout = {};
                if (metaData.$boxLayoutType == "100") {
                    metaData.$layout.$layoutType = self.boundFields.$allChildLayout[0].currentValue;
                }
                else {
                    metaData.$layout.$layoutType = "columns";
                    metaData.$layout.$layoutSubType = metaData.$boxLayoutType.replace(/-/g, ",");
                }
            }
            self.designedNode.item.applyDesignMetaData(metaData);
            if (metaData.$isHidden !== undefined) {
                self.designedNode.applyHiddenState(metaData);
            }
        }
        self.authoringView.isPageAuthored = true;
        self.applyChange(metaData);
        return null;
    },
    applyChange: function(newData){
        var self = this;
        var metaData = {
            $: {}
        };
        if (newData) {
            Object.keys(newData).forEach(function($prop){
                var $bind = $prop;
                if (self.designedNode.$authoringLevel == "field") {
                    switch ($prop) {
                        case "$isRightTextLabelAlignment":
                        case "$isTopLabelAlignment":
                        case "$isHidden":
                        case "$isTitleHidden":
                            $bind = "$field" + $prop;
                            break;
                    }
                }
                var value = newData[$prop];
                if ($bind == "$layout") {
                    $bind = "$boxLayoutType";
                    switch (newData.$layout.$layoutType || "stack") {
                        case "stack":
                        case "side":
                        case "tabs":
                            value = "100";
                            metaData.$allChildLayout = newData.$layout.$layoutType;
                            break;
                        default:
                            value = (newData.$layout.$layoutSubType || "50,50").replace(/,/g, "-");
                            break;
                    }
                }
                switch ($bind) {
                    case "$boxLayoutType":
                        if (!newData.$allChildLayout && !metaData.$allChildLayout) {
                            metaData.$allChildLayout = "stack";
                        }
                        metaData.$.$allChildLayout = {
                            $isDisabled: value != "100"
                        };
                        break;
                    case "$field$isTopLabelAlignment":
                        metaData.$.$field$isRightTextLabelAlignment = {
                            $isDisabled: value
                        };
                        break;
                    case "$field$isHidden":
                    case "$isHidden":
                        if (self.designedNode.$authoringLevel == "field") {
                            $bind = "$field$isHidden";
                        }
                        var $isDisabled = {
                            $isDisabled: value
                        };
                        Object.keys(self.$prototype.$).forEach(function($key){
                            if ($key.indexOf("$isHidden") < 0) {
                                metaData.$[$key] = $isDisabled;
                            }
                        });
                        if (self.menuItems) {
                            if (self.menuItems.$switchMode) {
                                self.menuItems.$switchMode[0].setMenu($isDisabled);
                                self.menuItems.$addItem[0].setMenu($isDisabled);
                            }
                        }
                        break;
                    case "$field$isTitleHidden":
                    case "$isTitleHidden":
                        metaData.$.$isBoxCollapsable = metaData.$.$isMaximizable = {
                            $isDisabled: value
                        };
                        break;
                }
                metaData[$bind] = value;
            });
        }
        Article.prototype.applyChange.call(this, metaData);
    },
    loadBox: function(){
        var self = this;
        if (self.$prototype.$field$isTopLabelAlignment = self.designedNode.item.getArticle().$isEditMode) {
            self.$prototype.$.$field$isRightTextLabelAlignment.$isDisabled = true;
        }
        Article.prototype.loadBox.call(this, this.designedNode.item.$item);
        if (self.menuItems.$switchMode) {
            self.menuItems.$switchMode[0].onMenuClick = function(){
                var node = self.designedNode;
                node.toggleLightMode();
                this.setTitle(node.$isLightMode ? _localize.fullMode : _localize.lightMode);
				if(node.$isLightMode){
					this.$$item[0].className = "s-author-designer-link s-author-designer-link-switch-" + node.$isLightMode?"light":"full"; 	
				}				
                return false;
            };
            self.menuItems.$addItem[0].onMenuClick = function(){
                _addNewBox(self.designedNode.item, self.$childCategory);
                return false;
            };
        }
    },
    dispose: function(){
        /* if (this.$$item) {
         this.$$item.remove();
         }*/
        delete this.authoringView;
        delete this.designedNode;
        Article.prototype.dispose.call(this);
    }
});

function FieldPalette(){
}

exports.FieldPalette = helpers.defineClass(FieldPalette, ItemPalette, {
    loadBox: function(){
        this.$prototype = {
            "$": {
                $field$isHidden: {
                    $type: "application/x-boolean",
                    $title: _localize.isHidden
                },
                $field$width: _formatFiewWidth(true),
                $field$isTopLabelAlignment: _formatFieldLabelAlign(true),
                $field$isTitleHidden: {
                    $type: "application/x-boolean",
                    $title: _localize.isLabelHidden
                },
                $field$isRightTextLabelAlignment: _formatFieldTextLabelAlign(true)
            },
            $field$width: "fit",
            $field$isTopLabelAlignment: false,
            $field$isTitleHidden: false,
            $field$isRightTextLabelAlignment: false
        };
        document.itemFactory.initializeItem(this, {
            $category: "section",
            $bind: "$field",
            $isSeparatorsVisible: true,
            $skin: "s-author-designer-item",
            $layout: {
                $layoutType: "columns",
                $layoutSubType: "50,50",
                $items: [{
                    $layoutType: "stack",
                    $items: [{
                        $category: "section",
                        $title: _localize.appearance,
                        $layout: {
                            $layoutType: "columns",
                            $layoutSubType: "50,50",
                            $items: [{
                                $layoutType: "stack",
                                $items: [_formatDefaultItem("$field$isHidden"), _formatFiewWidth(false)]
                            }, {
                                $layoutType: "stack",
                                $items: [_formatDefaultItem("$field$isTitleHidden"), _formatFieldsLabelPositionBlock()]
                            }]
                        }
                    }]
                }, {
                    $layoutType: "stack",
                    $items: [{
                        $category: "slot",
                        $id: "s-author-designer-custom-field"
                    }]
                }]
            
            }
        }, this.authoringView);
        ItemPalette.prototype.loadBox.call(this);
        if (this.designedNode.item.getAuthoringWidget) {
            if (this.fieldAuthoring = this.designedNode.item.getAuthoringWidget()) {
                this.fieldAuthoring.$$container = $("#s-author-designer-custom-field");
                document.itemFactory.initializeItem(this.fieldAuthoring, {}, this);
                this.fieldAuthoring.designedField = this.designedNode.item;
                this.fieldAuthoring.$layoutOptions = document.site.$item.$layoutOptions.authoring;
                this.fieldAuthoring.$skin = this.fieldAuthoring.$layoutOptions.section;
                this.fieldAuthoring.loadBox(this.designedNode.item.$item);
            }
        }
    },
    dispose: function(){
        if (this.fieldAuthoring) {
            delete this.fieldAuthoring.designedField;
            delete this.fieldAuthoring;
        }
        ItemPalette.prototype.dispose.call(this);
    }
});

function PagePalette(){
}

exports.PagePalette = helpers.defineClass(PagePalette, ItemPalette, {
    loadBox: function(){
        this.$childCategory = "section";
        this.$prototype = {
            "$": {
                $isSeparatorsVisible: {
                    $type: "application/x-boolean",
                    $title: _localize.isSeparatorsVisible
                },
                $boxLayoutType: _formatLayout(true),
                $allChildLayout: _formatChildLayout(true, this.$childCategory),
                $field$width: _formatFiewWidth(true),
                $field$isTopLabelAlignment: _formatFieldLabelAlign(true),
                $field$isRightTextLabelAlignment: _formatFieldTextLabelAlign(true)
            },
            $links: _formatLinks(_localize.addSection),
            $boxLayoutType: "100",
            $isSeparatorsVisible: false,
            $field$width: "fit",
            $field$isTopLabelAlignment: false,
            $field$isRightTextLabelAlignment: false
        };
        document.itemFactory.initializeItem(this, {
            $category: "section",
            $bind: "$box",
            $isSeparatorsVisible: true,
            $skin: "s-author-designer-item",
            $layout: {
                $layoutType: "columns",
                $layoutSubType: "60,20,20",
                $items: [{
                    $layoutType: "stack",
                    $items: [{
                        $layoutType: "columns",
                        $layoutSubType: "50,50",
                        $items: [{
                            $layoutType: "stack",
                            $items: [_formatLayout(false)]
                        }, {
                            $layoutType: "stack",
                            $items: [{
                                $category: "section",
                                $title: _localize.appearance,
                                $layout: {
                                    $items: [{
                                        $bind: "$isSeparatorsVisible",
                                        $skin: "s-author-field",
                                        $isEditMode: true,
                                        $isLeftSpaceHidden: true
                                    }]
                                }
                            }]
                        }]
                    }, _formatLinks()]
                }, {
                    $layoutType: "stack",
                    $items: [_formatChildLayout(false, this.$childCategory)]
                }, {
                    $layoutType: "stack",
                    $items: [{
                        $category: "section",
                        $bind: "$field",
                        $title: _localize.allFields,
                        $layout: {
                            $items: [_formatFiewWidth(false), _formatFieldsLabelPositionBlock()]
                        }
                    }]
                }]
            
            }
        }, this.authoringView);
        ItemPalette.prototype.loadBox.call(this);
    }
    
});


function SectionPalette(){
}

exports.SectionPalette = helpers.defineClass(SectionPalette, ItemPalette, {
    loadBox: function(){
        this.$childCategory = "block";
        this.$prototype = {
            "$": {
                $isHidden: {
                    $type: "application/x-boolean",
                    $title: _localize.isHidden
                },
                $isMaximizable: {
                    $type: "application/x-boolean",
                    $title: _localize.isMaximizable
                },
                $isBoxCollapsable: {
                    $type: "application/x-boolean",
                    $title: _localize.isBoxCollapsable
                },
                $isTitleHidden: {
                    $type: "application/x-boolean",
                    $title: _localize.isTitleHidden
                },
                $isSeparatorsVisible: {
                    $type: "application/x-boolean",
                    $title: _localize.isSeparatorsVisible
                },
                $boxLayoutType: _formatLayout(true),
                $allChildLayout: _formatChildLayout(true, this.$childCategory),
                $field$width: _formatFiewWidth(true),
                $field$isTopLabelAlignment: _formatFieldLabelAlign(true),
                $field$isRightTextLabelAlignment: _formatFieldTextLabelAlign(true)
            },
            $links: _formatLinks(_localize.addBlock),
            $isSeparatorsVisible: false,
            $boxLayoutType: "100",
            $field$width: "fit",
            $field$isTopLabelAlignment: false,
            $field$isRightTextLabelAlignment: false
        };
        var designedItem = this.designedNode.item;
        document.itemFactory.initializeItem(this, {
            $category: "section",
            $bind: "$box",
            $isSeparatorsVisible: true,
            $skin: "s-author-designer-item",
            $layout: {
                $layoutType: "columns",
                $layoutSubType: "60,15,25",
                $items: [_formatBoxLayoutStack(), {
                    $layoutType: "stack",
                    $items: [_formatChildLayout(false, this.$childCategory)]
                }, {
                    $layoutType: "stack",
                    $items: [{
                        $category: "section",
                        $bind: "$field",
                        $title: _localize.allFields,
                        $layout: {
                            $items: [_formatFiewWidth(false), _formatFieldsLabelPositionBlock()]
                        }
                    }]
                }]
            
            }
        }, this.authoringView);
        ItemPalette.prototype.loadBox.call(this);
    }
});
function BlockPalette(){
}

exports.BlockPalette = helpers.defineClass(BlockPalette, ItemPalette, {
    loadBox: function(){
        this.$childCategory = "field";
        this.$prototype = {
            "$": {
                $isHidden: {
                    $type: "application/x-boolean",
                    $title: _localize.isHidden
                },
                $isMaximizable: {
                    $type: "application/x-boolean",
                    $title: _localize.isMaximizable
                },
                $isBoxCollapsable: {
                    $type: "application/x-boolean",
                    $title: _localize.isBoxCollapsable
                },
                $isTitleHidden: {
                    $type: "application/x-boolean",
                    $title: _localize.isTitleHidden
                },
                $isSeparatorsVisible: {
                    $type: "application/x-boolean",
                    $title: _localize.isSeparatorsVisible
                },
                $boxLayoutType: _formatLayout(true),
                $allChildLayout: _formatChildLayout(true, this.$childCategory),
                $field$width: _formatFiewWidth(true),
                $field$isTitleHidden: {
                    $type: "application/x-boolean",
                    $title: _localize.isLabelHidden
                },
                $field$isTopLabelAlignment: _formatFieldLabelAlign(true),
                $field$isRightTextLabelAlignment: _formatFieldTextLabelAlign(true)
            },
            $links: _formatLinks(_localize.addField),
            $isSeparatorsVisible: false,
            $boxLayoutType: "100",
            $field$width: "fit",
            $field$isTopLabelAlignment: false,
            $field$isRightTextLabelAlignment: false
        };
        var designedItem = this.designedNode.item;
        document.itemFactory.initializeItem(this, {
            $category: "section",
            $bind: "$box",
            $isSeparatorsVisible: true,
            $skin: "s-author-designer-item",
            $layout: {
                $layoutType: "columns",
                $layoutSubType: "60,40",
                $items: [_formatBoxLayoutStack(), {
                    $layoutType: "stack",
                    $items: [{
                        $category: "section",
                        $bind: "$field",
                        $title: _localize.allFields,
                        $layout: {
                            $layoutType: "columns",
                            $layoutSubType: "50,50",
                            $items: [{
                                $layoutType: "stack",
                                $items: [_formatChildLayout(false, this.$childCategory), _formatFiewWidth(false)]
                            }, {
                                $layoutType: "stack",
                                $items: [_formatDefaultItem("$field$isTitleHidden"), _formatFieldsLabelPositionBlock()]
                            }]
                        }
                    }]
                }]
            
            }
        }, this.authoringView);
        ItemPalette.prototype.loadBox.call(this);
    }
});
