// syracuse-ui/lib/desktop/site/authoring/treeNode.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function TreeNode(item, parentNode, authoringView){
    this.item = item;
    this.$isExpanded = false;
    this.children = [];
    this.parentNode = parentNode;
    this.authoringView = authoringView;
}

exports.TreeNode = helpers.defineClass(TreeNode, null, {
    getChildrenLevel: function(){
        return this.children.length > 0 ? this.children[0].$authoringLevel : null;
    },
    setDesignableItem: function(item){
        if (item.$isTabLayout) {
            item.$$header.attr("s-designable", "1").attr("data-s-item", item.id);
            item.$$header[0].style.border = "2px dashed transparent";
			item.$$body[0].style.border = "2px dashed transparent";
            if (item.boxParent) {
                item.$$header.attr("data-s-article", item.boxParent.getArticle().id);
            }
        }
        item.$$item.attr("s-designable", "1").attr("data-s-item", item.id);
        if (item.boxParent) {
            item.$$item.attr("data-s-article", item.boxParent.getArticle().id);
        }
        (this.$$dataSlot || item.$$item)[0].style.border = "2px dashed transparent";
    },
    selectItem: function(doSelect){
        this.authoringView.selectItem(this, doSelect);
    },
    drawChildNodes: function(draw){
        this.children.forEach(function(child){
            child.drawNode(draw);
        });
    },
    addNodeId: function(target){
        target.setAttribute("data-s-node", this.id);
    },
    drawNode: function(draw){
        if (draw) {
            var dom;
            if (!this.parentNode.$$children) {
                this.parentNode.$$opener[0].className = "s-author-tree-item-opener-children";
                dom = document.createElement("ul");
                dom.className = "s-author-tree-level";
                this.parentNode.$$children = $(dom).appendTo(this.parentNode.$$node);
                this.parentNode.expand(false);
            }
            
            dom = document.createElement("li");
            dom.className = "s-author-tree-item";
            this.$$node = $(dom).appendTo(this.parentNode.$$children);
            dom = document.createElement("div");
            dom.className = "s-author-tree-item-data";
            this.$$data = $(dom).appendTo(this.$$node);
            
            dom = document.createElement("a");
            dom.className = "s-author-tree-item-opener";
            this.addNodeId(dom);
            this.$$opener = $(dom).appendTo(this.$$data);
            
            var check = (this.$$check = $("<input type='checkbox'/>").appendTo(this.$$data))[0];
            check.className = "s-author-tree-item-check";
            this.addNodeId(check);
            this.applyHiddenState(this);
            
            dom = document.createElement("div");
            dom.className = "s-author-tree-item-title" + (this.authoringView.$$pathSteps[this.id] ? (" s-author-tree-designed-" + this.$authoringLevel) : "");
            this.addNodeId(dom);
            this.$$title = $(dom).text(this.item.getTitle()).appendTo(this.$$data);
        }
        else {
            delete this.$$node;
            delete this.$$data;
            delete this.$$opener;
            delete this.$$check;
            delete this.$$title;
            delete this.$$children;
        }
        this.drawChildNodes(draw);
    },
    load: function(){
        var self = this;
        self.id = document.controller.generateUUID();
        self.authoringView.nodes[self.id] = self;
        self.$authoringLevel = self.item.$authoringLevel || "field";
        
        self.item.authoringNode = self;
        if (self.item.$$item) {
            if (!self.parentNode) {
                self.$$dataSlot = self.item.$$dataSlot; //for page
                self.item.$$item.bind("click.treenoderoot", function(event){
                    if (!document.site.DDAuthoring) {
                        self.selectItem();
                    }
                    return false;
                });
            }
            self.setDesignableItem(self.item);
            if (self.item.layoutContent) {
                self.item.layoutContent.toggleAuthoring(true);
            }
        }
        self.applyHiddenState(self.item.$item);
        //self.draw();
        self._loadLayout(self.item.layoutContent);
        return self;
    },
    _loadLayout: function(layoutContent){
        if (layoutContent) {
            var self = this;
            layoutContent.items.forEach(function(item){
                if (item.$layout) {
                    self._loadLayout(item);
                }
                else {
                    self.appendChildItem({
                        newItem: item
                    });
                }
            });
        }
    },
    toggleLightMode: function(layoutContent, $isLightMode){
        var self = this;
        if (!layoutContent) {
            layoutContent = self.item.layoutContent;
        }
        self.$isLightMode = ($isLightMode !== undefined) ? $isLightMode : !self.$isLightMode;
        if (layoutContent) {
            layoutContent.items.forEach(function(child){
                if (child.$layout) {
                    self.toggleLightMode(child, self.$isLightMode);
                }
                else {
                    var node = child.authoringNode;
                    if (node.$authoringLevel == "field") {
                        if (self.$isLightMode) {
                            child.authoringLight = {
                                $$bag: $(document.createElement("div")).append(child.$$item.children()),
                                $className: child.$$item[0].className
                            }
                            child.$$item[0].className = "s-author-light-field s-authoring-designable"
                            child.$$item[0].setAttribute("s-designable", "1");
                            
                            var div = document.createElement("div");
                            div.className = "s-author-light-field-value";
                            var $$value = $(div);
                            
                            var div = document.createElement("div");
                            div.className = "s-author-light-field-type";
                            div.style.backgroundImage = "url('" + document.site.$item.$iconPath + "authoring/s-author-light-" + child.$field.$type.replace("application/x-", "") + ".png')";
                            $$value.append(div);
                            
                            div = document.createElement("div");
                            div.className = "s-author-light-field-title";
                            $$value.append($(div).text(child.getTitle()));
                            $$value.appendTo(child.$$item);
                        }
                        else {
                            if (child.authoringLight) {
                                child.$$item.empty()[0].className = child.authoringLight.$className;
                                child.$$item.append(child.authoringLight.$$bag.children());
                                delete child.authoringLight;
                            }
                        }
                    }
                    node.toggleLightMode(null, self.$isLightMode);
                }
            });
        }
    },
    applyHiddenState: function(state){
        this.$isHidden = state.$isHidden;
        if (this.$$check) {
            this.$$check.attr("checked", this.$isHidden !== true);
        }
    },
    toggleOverItem: function(isOver){
        this.toggleItemCss("over", isOver);
        if (isOver && this.authoringView.overNode != this) {
            this.authoringView.showOverPathNodes(this, true);
        }
    },
    toggleItemCss: function(status, show, css){
        css = css || ("s-author-" + status + "-" + this.$authoringLevel);
        if (this.item && this.item.$$item) {
            var $$item = this.$$dataSlot || this.item.$$item;
            document.site.toggleClass($$item, css, show);
            $$item[0].style.borderColor = show ? "" : "transparent";
            if (this.item.$isTabLayout) {
                this.item.$$header[0].style.borderColor = show ? "" : "transparent";
                document.site.toggleClass(this.item.$$header, css, show);
            }
        }
    },
    highlightDesignedItem: function($designed){
        this.toggleItemCss("designed", $designed);
    },
    appendChildItem: function(options){
        var child = options.newItem.authoringNode;
        if (!child) {
            child = options.newItem.authoringNode = new TreeNode(options.newItem, this, this.authoringView);
        }
        else {
            if (child.parentNode != this) {
                child.parentNode.removeChildNode(child);
                //child.parentNode.children.splice(child.parentNode.getNodeIndex(child), 1);
            }
        }
        this.children.splice(options.newIndex || this.children.length + 1, 0, child);
        child.parentNode = this;
        if (options.load !== false) {
            child.load();
        }
    },
    removeChildNode: function(child){
        this.children.splice(this.getNodeIndex(child), 1);
    },
    getNodeIndex: function(node){
        for (var ii = 0; ii < this.children.length; ii++) {
            if (this.children[ii] == node) {
                return ii;
            }
        }
        return -1;
    },
    
    expand: function($isExpanded){
        this.$isExpanded = $isExpanded === undefined ? !this.$isExpanded : $isExpanded;
        if (this.$$opener) {
            this.$$opener.toggleClass("s-open", this.$isExpanded);
        }
        if (this.$$children) {
            this.$$children.toggle(this.$isExpanded);
        }
    },
    isPageItem: function(){
        return this.item.getPage ? (this.item.getPage() == this.item) : false;
    },
    dispose: function(){
        if (this.item) {
            if (this.item.$$item && !this.parentNode) {
                this.item.$$item.unbind("click.treenoderoot");
            }
            this.toggleLightMode(null, false);
            delete this.item.authoringNode;
            if (this.item.layoutContent) {
                this.item.layoutContent.toggleAuthoring(false);
            }
            this.highlightDesignedItem(false);
            delete this.item;
        }
        if (this.children) {
            this.children.forEach(function(child){
                child.dispose();
            });
        }
        this.parentNode = this.authoringView = this.children = null;
    }
});
