// syracuse-ui/lib/field/reference/lookup.js
"use strict";
var List = require('syracuse-ui/lib/scroll/list').List;

function Lookup(field) {
	this.field = field;
	this.id = field.id + "-lookup";
	syra_item.register(this);
	if ((field.articleParent && field.articleParent.onLookupClick) ? field.articleParent.onLookupClick(field) : true) {
		field.focus && field.focus();
		var $lookup = field.$menus.$lookup;
		if (field.isChildFieldRecord) {
			field.articleParent.currentSelectRecords = null;
		}
		this.fetch();
	}
}

Lookup.prototype.fetch = function() {
	var self = this;
	if (self.field.$item.$summary) {
		self.showSummary();
	} else {
		syra_over.openModal(self.field.boxParent, {
			article: self.field.articleParent,
			$url: self.field.$menus.$lookup.$url,
			onSelectRecord: function(records) {
				self.select(records[Object.keys(records)[0]].dataset);
			}
		});
	}
};
Lookup.prototype.click = function(event, target) {
	var self = this;
	if (target.syraManage) {
		self._modalOpened = true;
		self.page.startChange();
		self.page.applyChange(self.data);
		self.page.endChange();
		self.page.menuBar && self.page.menuBar.toggleBar(false);
		syra_over.openModal(self.field.boxParent, {
			page: self.page,
			onSelectRecord: function(records) {
				self.select(records[Object.keys(records)[0]].dataset);
			},
			close: function() {
				self._modalOpened = false;
				setTimeout(function() {
					self.dispose();
				}, 100);
			}
		});
	} else {
		if (target.syraRecord !== undefined) {
			self.select(self.data[self.$arrayBind][target.syraRecord]);
			self.dispose();
		}
	}

};
Lookup.prototype.showSummary = function() {
	var self = this;
	self.isLoading = true;
	syra_router.loadRepresentation({
		article: self.field.articleParent,
		segments: self.field.$menus.$lookup.$url,
		success: function success($itemPage) {
			syra_pageBuilder.load({
				$itemPage: $itemPage,
				$autoFetch: false,
				success: function success(page) {
					if (self.field) {
						self.page = page;
						var selected = self.field.getValue();
						if (selected && typeof(selected) == "object") {
							selected = selected.$uuid;
						} else {
							selected = null;
						}
						var $properties = self.page.$prototype.$properties;
						var $binds = Object.keys($properties);

						for (var ii = 0, jj = $binds.length; ii < jj; ii++) {
							if ($properties[$binds[ii]].$type == "application/x-array") {
								self.$arrayBind = $binds[ii];
								break;
							}
						}

						self.page.fetch(null, function(data, response, requestUrl) {
							if (self.field) {
								var slot = syra_dom.div();
								self.data = data;
								self.list = new List({
									css: "s-list-primary",
									parentNode: slot,
									scroll: true
								});
								var records = data[self.$arrayBind];
								var link;
								for (var ii = 0, jj = records.length; ii < jj; ii++) {
									var record = records[ii];
									link = self.list.addAnchor();
									link.textContent = record.$value;
									link.syraRecord = ii;
									link.syraItem = self.id;
									link.syraOnClick = "click";
									if (selected == record.$uuid) {
										var li = link.parentNode;
										li.parentNode.insertBefore(li, li.parentNode.firstChild);
										syra_dom.toggleClass(link, "s-selected", true);
										syra_dom.toggleClass(li, "s-selected", true);
									}
								}

								self.detailsLink = link = syra_dom.anchor("s-list-primary-top-sep s-list-primary-btn-primary", slot);
								var detailText = self.field.$item.$summary.$details || syra_local.fieldSummaryDetail.replace("{0}", self.field.getTitle());

								if (data.$totalResults && data.$totalResults != records.length) {
									detailText += (" ({0}/{1})").replace("{0}", records.length).replace("{1}", data.$totalResults);
								}
								link.textContent = detailText;
								link.syraManage = true;
								link.syraItem = self.id;
								link.syraOnClick = "click"

								var position = self.field.$item.$popupRight ? {
									my: "left top",
									at: "right top",
									of: self.field.domItem
								} : {
									my: "left top",
									at: "left bottom",
									of: self.field._dataValue
								};
								self._popup = syra_over.openPopup(syra_site, {
									slot: slot,
									//autoCloseBoundary: self.field.domItem,
									picker: (self.field.$item.$picker == "field") ? self.field.domItem : self.field._dataValue,
									position: position,
									onresize: function() {
										if (self.list) {
											self.list.resize(syra_site.size.popupMaxHeight - self.detailsLink.clientHeight);
										}
									},
									close: function() {
										self.list && self.list.dispose();
										self._popup = null;
										!self._modalOpened &&
											setTimeout(function() {
												self.dispose();
											}, 100);

									}
								});
								self.isLoading = false;
							}
						});
					} else {
						page.dispose();
					}
				}
			});
		}
	});
};

Lookup.prototype.select = function(dataset) {
	if (!this.field.$isDisabled && !this.field.$isReadOnly) {
		if (this.field.isChildFieldRecord) {
			this.field.articleParent.setValue(dataset);
			syra_form.updateDelta(this.field.articleParent, dataset);
		} else {
			this.field.setValue(this.field.$menus.$lookup.$result ? dataset[this.field.$menus.$lookup.$result] : dataset);
			this.field.focus();
			syra_form.update(this.field, this.field.currentValue);
		}
	}
};
Lookup.prototype.dispose = function() {
	if (this.field) {
		delete this.field._lookupItem;
	}
	this._popup && this._popup.close();
	this.page && this.page.dispose();
	syra_item.unregister(this);
	syra_article.dispose(this);
};

function _clickPicker(options, event) {
	var field = this.parent;
	if (field.$item.$openBy == "over" && !this.delayActivated) {
		return;
	}
	if (field._lookupItem) {
		if (field._lookupItem.isLoading) {
			return;
		}
		field._lookupItem.dispose();
	}
	field._lookupItem = new Lookup(field);
}

exports.addPicker = function(field, $menu) {
	if (field.$isEditMode && !field.picker_$lookup) {
		var btn = {
			parent: field,
			pickerId: "$lookup",
			fontIcon: "lookup",
			text: syra_local.fieldLookup,
			shortCutTip: !field.$item.$isShortcutDisabled && syra_shortCuts.tip.openLookup,
			isHidden: field.$item.$isPickerHidden,
			click: _clickPicker
		};
		if (field.$item.$summary) {
			btn.fontIcon = field.$item.$popupRight ? "right_arrow" : "expand_m";
			btn.targetClick = field._dataValue;
		}
		syra_button.addFieldPicker(btn);
	}
	$menu.$isPopup = false;
};
