// syracuse-ui/lib/menus/clickHandlers.js
"use strict";
var _dynamicMenu;

function _clearDynamicMenu() {
	_dynamicMenu && _dynamicMenu.articleParent && _dynamicMenu.articleParent.removeItem(_dynamicMenu, true, true);
	_dynamicMenu = null;
}

exports.main = function(page) {
	var $binds = Object.keys(page.menuItems);
	for (var ii = 0, jj = $binds.length; ii < jj; ii++) {
		var menus = page.menuItems[$binds[ii]];
		if (menus && menus.length && menus[0].$item.$style === "main") {
			exports.menuItem(menus[0]);
			break;
		}
	}
};

exports.save = function(field) {
	return exports.menuId(field, "$save");
};


exports.prev = function(parent, toFirst) {
	return exports.menuId(parent, toFirst ? "$first" : "$previous");
};

exports.next = function(parent, toLast) {
	return exports.menuId(parent, toLast ? "$last" : "$next");
};

exports.navigation = function() {
	return exports.menuId(syra_site, "$navigation");
};

exports.home = function() {
	return exports.menuId(syra_site, "$home");
};

exports.detail = function(field) {
	return exports.menuId(field, "$detail");
};


exports.create = function(field) {
	return exports.menuId(field, field.isFusionPage ? "1115" : "$create");
};

exports.contextualMenu = function(field) {
	if (field.mnPickers && field.mnPickers.menus) {
		setTimeout(function() {
			field.mnPickers.menus.click();
		}, 10);
		return true;
	}
};

exports.selector = function(field, dateTimeOnly) {
	var pickers = field.mnPickers;
	if (pickers) {
		var picker = (!dateTimeOnly && (pickers.$lookup || pickers.choice)) || pickers.date || pickers.time;
		if (picker) {
			setTimeout(function() {
				picker.click();
			}, 10);
			return true;
		}
	}
};

exports.tunnel = function(field) {
	if (field.mnPickers && field.mnPickers.$tunnel) {
		setTimeout(function() {
			field.mnPickers.$tunnel.click();
		}, 10);
	} else {
		if (field.$menusPopup && field.$menusPopup.$tunnel) {
			field.page.externalAdapter.onFieldClickPicker({
				field: field,
				pickerType: "$tunnel"
			});
		}
	}
};

exports.print = function(page) {
	return exports.menuId(page, "$print");
};

exports.deleteItem = function(item) {
	return exports.menuId(item, "$delete");
};

exports.refreshPageData = function(page) {
	if (page.isFusionPage) {
		exports.menuId(page, "2823");
	} else {
		var landingPage;
		if (page.isLandingMaster) {
			landingPage = page.getSelectedChildPage();
		} else {
			landingPage = page.isLandingPage ? page : null;
		}
		if (landingPage) {
			landingPage.refreshPicker.click();
		} else {
			page.$autoFetch &&
				page.fetch({}, function(data, response, requestUrl) {
					if (!page || !page.disposed) {
						page.startChange();
						page.applyChange(data, response, requestUrl);
						page.endChange();
					}
				});
		}
	}
};


exports.showFieldProperty = function(field) {
	if (field.page.isFusionPage) {
		exports.menuId(field.page, "1172");
	} else {

	}
};

exports.goToSearch = function(page) {
	if (page.isFusionPage) {
		exports.menuId(page, "1186");
	} else {
		if (syra_site.searcher) {
			syra_site.searcher.setFocusToSearchField();
			return true;
		}
	}
};


function _onClick(menu, event) {
	var doClick = true;
	if (doClick && menu.contextField) {
		doClick = menu.contextField.onMenuClick ? menu.contextField.onMenuClick(menu, event) : true;
	}
	if (doClick && menu.$item.$ownerId) {
		var owner = syra_store.get(menu.$item.$ownerId);
		doClick = owner && owner.onMenuClick ? owner.onMenuClick(menu, event) : true;
	}

	if (doClick && menu.boxParent) {
		doClick = menu.boxParent.onMenuClick ? menu.boxParent.onMenuClick(menu, event) : true;
		if (doClick && (menu.boxParent != menu.articleParent)) {
			doClick = menu.articleParent.onMenuClick ? menu.articleParent.onMenuClick(menu, event) : true;
		}
	}

	if (doClick) {
		if (menu.articleParent != menu.page) {
			doClick = menu.page.onMenuClick ? menu.page.onMenuClick(menu, event) : true;
		}
		if (doClick && (menu.onMenuClick ? menu.onMenuClick(event) : true)) {
			syra_controller.executeMenu(menu);
		}
	}
	setTimeout(function() {
		if (menu) {
			var parent = menu.boxParent;
			while (parent && !parent.disposed) {
				if (parent.menusPopup) {
					parent.menusPopup.close();
					break;
				}
				parent = parent.boxParent;
			}
		}
	}, 10);
}


exports.dynamicMenu = function(article, $link) {
	var $bind = article.id + "-syra-dynamic-menu";
	article.$menus[$bind] = null;
	var $links = {};
	$links[$bind] = $link;
	article.applyChange({
		$links: $links
	});
	return exports.menuId(article, $bind);
};
exports.menuId = function(field, $bind) {
	var menuItem = field.menuItems && field.menuItems[$bind];
	if (!menuItem) {
		var $menu = field.$menus[$bind];
		if ($menu) {
			var article = (field.isDesktopPage || field.isRecordArticle) ? field : field.articleParent;
			_clearDynamicMenu();
			menuItem = _dynamicMenu = field.page.loadNewItem(null, {
				$category: "link",
				$bind: $menu.$bind,
				$isAction: $menu.$isAction
			}, article);
			menuItem.contextField = field;
		}
	} else {
		menuItem = menuItem[0];
	}
	if (menuItem) {
		exports.menuItem(menuItem);
		return true;
	}
};



exports.menuItem = function(menu, event) {
	if (!menu.$isDisabled) {
		if (menu.page) {
			if (menu.isNoSyracuseUrl) {
				if (event) {
					event.syraRetValue = true;
				} else {
					window.open(menu.$url, "_blank");
				}
				return;
			}
			if (menu.page.checkMenuDesignValidaty(menu)) {
				var isValidated = true;
				if (menu.$bind == "$save" || menu.$bind == "$saveAs") {
					isValidated = menu.page.validateFields();
				}
				if (isValidated) {
					var field;
					if (menu.boxParent && menu.boxParent.contextField) {
						//boxParent is menubox of contextMenu
						menu.contextField = field = menu.boxParent.contextField;
						if (menu.contextField && menu.contextField.$item && !menu.$subRecordKey) {
							menu.$subRecordKey = menu.contextField.$item.$bind;
						}
					}
					var page = menu.page;
					/*var page = menu.page.mainPage,
						dialog = syra_site.dialogManager.getTopDialogPage();
					if (dialog && dialog._content && !dialog._content.disposed) {
						page = dialog._content;
					}
					if (!page || page.disposed) {
						page = menu.page;
					}*/
					if (page && !page.disposed && page.externalAdapter) {
						page.externalAdapter.onMenuItemClick({
							menuItem: menu,
							field: field,
							doEvent: function() {
								_onClick(menu, event);
							}
						});
					}
				}
			}
		} else {
			_onClick(menu, event);
		}
	}
};


exports.dispose = function() {
	_clearDynamicMenu();
};
