// syracuse-ui/lib/menus/menusHelper.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var shortcutInvoker = require("syracuse-ui/lib/events/shortCutInvoker");
var _embeddedHelper = require('syracuse-ui/lib/field/helpers/embeddedHelper');
exports.click = require('./clickHandlers');
var _pagingCapability = require("syracuse-ui/lib/field/array/capability/pagingCapability");

exports.btnWidth = 20;

var _fontIcons = {
	filter_empty: "&#xf100;",
	details: "&#xf101;",
	escape: "&#xf103;",
	bookmark_off: "&#xf104;",
	menus: "&#xf105;",
	previous: "&#xf106;",
	info: "&#xf107;",
	filter_gt: "&#xf109;",
	field_time: "&#xf10a;",
	time: "&#xf10a;",
	select: "&#xf102;",
	field_choice: "&#xf10b;",
	refresh: "&#xf10d;",
	resume: "&#xf10e;",
	next: "&#xf10e;",
	role: "&#xf110;",
	checkbox_off: "&#xf111;",
	help: "&#xf11e;",
	checkbox_on: "&#xf113;",
	field_boolean: "&#xf113;",
	field_datetime: "&#xf114;",
	field_date: "&#xf114;",
	date: "&#xf114;",
	menus_open: "&#xf115;",
	endpoint: "&#xf116;",
	theme: "&#xf117;",
	show_advanced: "&#xf118;",
	external: "&#xf11a;",
	radio_on: "&#xf11b;",
	collapse_row: "&#xf11c;",
	email: "&#xf11f;",
	hide_advanced: "&#xf122;",
	field_number: "&#xf123;",
	node_expanded: "&#xf125;",
	node_collapsed: "&#xf127;",
	right_arrow: "&#xf127;",
	next_step: "&#xf127;",
	radio_off: "&#xf128;",
	expand_row: "&#xf129;",
	left_arrow: "&#xf12d;",
	prev_step: "&#xf12d;",
	quickEdit: "&#xf12e;",
	field_graph: "&#xf12f;",
	bookmark_on: "&#xf130;",
	maximize: "&#xf131;",
	expand: "&#xf132;",
	reload_next: "&#xf133;",
	settings: "&#xf134;",
	design: "&#xf134;",
	abort: "&#xf135;",
	cancel: "&#xf135;",
	close: "&#xf135;",
	boolean_off: "&#xf135;",
	tunnel: "&#xf137;",
	diagnose: "&#xf139;",
	last_step: "&#xf13b;",
	download: "&#xf13c;",
	upload: "&#xf13c;",
	first_step: "&#xf13d;",
	field_image: "&#xf13e;",
	remove_separator: "&#xf13f;",
	reload_back: "&#xf141;",
	filter_flag: "&#xf142;",
	filter_none: "&#xf142;",
	field_text: "&#xf143;",
	choice: "&#xf144;",
	lookup: "&#xf145;",
	search: "&#xf145;",
	first: "&#xf146;",
	"delete": "&#xf147;",
	last: "&#xf148;",
	query: "&#xf14a",
	field_array: "&#xf14a",
	phone: "&#xf14b;",
	localize: "&#xf14c;",
	locale: "&#xf14e;",
	save: "&#xf14f;",
	boolean_on: "&#xf14f;",
	suspend: "&#xf151;",
	add_separator: "&#xf151;",
	mandatory: "&#xf152;",
	field_string: "&#xf153;",
	create: "&#xf157;",
	minimize: "&#xf159;",
	edit: "&#xf15a;",
	collapse: "&#xf15b;",
	filter_like: "&#xf119;",
	filter_like_s: "&#xf120;",
	filter_between: "&#xf121;",
	filter_ge: "&#xf12b;",
	filter_lt: "&#xf138;",
	filter_notempty: "&#xf140;",
	filter_le: "&#xf156;",
	filter_eq: "&#xf13a;",
	filter_ne: "&#xf158;"
};


exports.cssFontIconBtn = " s-fonticon-btn ";
var _cssFontIconMn = " s-fonticon-mn ";

if (document.dir === 'rtl') {
	exports.cssFontIconBtn += " s-rtl-img ";
	_cssFontIconMn += " s-rtl-img ";
}

exports.fontIcons = _fontIcons;

function _setTitleIconValue(menu, $title, $iconValue) {
	if ($iconValue != undefined) {
		menu.$iconValue = $iconValue;
	}
	var name = menu.$iconValue || menu.value || menu.$item.$icon.$value;
	if (!name) {
		name = menu.$sourceBind || menu.$item.$bind;
	}
	name = (name[0] == "$") ? name.substr(1) : name;
	var fontIcon = syra_menus.fontIcons[name];

	if (menu.$item.$icon && !menu.iconValue) {
		menu.cssTitle = null;
		var cssBtn = _cssFontIconMn + "s-btn-" + name;
		if (menu.$item.$icon.$mode == "icon") {
			menu.$item.$noText = true;
			menu.iconValue = menu.domItem;
			menu.cssTitle = fontIcon ? cssBtn : " s-mn-icon";
		} else {
			menu.iconValue = document.createElement("div");
			var iconCss = menu.$skin + "-icon " + (fontIcon ? cssBtn : " s-mn-icon");
			if (menu.$item.$css) {
				iconCss += " " + menu.$item.$css;
			}
			menu.iconValue.className = iconCss;
			menu.cssTitle = " s-mn-mode-icontext";
		}
		menu.cssUpdated = true;
	}
	if (fontIcon) {
		menu.iconValue.innerHTML = fontIcon;
	} else {
		var $path = syra_site.$item.$iconPath + (menu.$item.$icon.$path || "");
		menu.iconValue.style.backgroundImage = "url('" + $path + name + ".png')";
	}
	if (menu.iconValue != menu.domItem) {
		menu.domItem.insertBefore(menu.iconValue, menu.domItem.firstChild);
	}
}

exports.setMenuItemTitle = function(menu, $title, $iconValue) {
	menu.$title = ($title || "") + ""; //ensure title is string
	delete menu.domItem.syraTitle;
	var title = menu.getTitle();
	menu.domItem.title = menu.$description || title;

	if (!menu.$item.$noText && !(menu.$item.$icon && menu.$item.$icon.$mode == "icon")) {
		if (title.indexOf("<em") >= 0) {
			menu.domItem.innerHTML = title.replace(/<em>/g, "<em class='" + menu.$skin + "-em'>");
		} else {
			menu.domItem.textContent = title;
		}
	}
	menu.$item.$icon && _setTitleIconValue(menu, $title, $iconValue);

	if (menu.$description && menu.$item.$isDescriptionVisible) {
		if (!menu._description) {
			menu._description = document.createElement("div");
			menu._description.className = menu.$item.$skinDescription || (menu.$skin + "-desc");
		}
		menu._description.textContent = menu.$description;
		menu.domItem.appendChild(menu._description);
	}
	if (menu.$target == "blank") {
		if (!menu.external) {
			menu.external = exports.addFontIconFlag("", "external");
		}
		menu.domItem.appendChild(menu.external);
	}
};

exports.addTextButton = function(text, css, onClick, shortCutTip, fontIcon) {
	var btn = document.createElement("a");
	btn.className = css;
	btn.setAttribute("href", "#"); //oneed to improve cancel action otherwiese add # to url
	btn.syraOnClick = onClick;
	btn.textContent = btn.title = text;
	btn.syraShortCutTip = shortCutTip;
	btn.syraTip = 1;
	if (fontIcon) {
		btn.syraHasIcon = true;
		btn.insertBefore(exports.addFontIconFlag(css + "-flag", fontIcon), btn.firstChild);
	}
	return btn;
};

function _setFontIconButton(btn, fontIcon) {
	if (fontIcon[0] == "$") {
		fontIcon = fontIcon.substr(1);
	}
	var icon = _fontIcons[fontIcon];
	if (icon) {
		btn.innerHTML = icon;
		btn.className = btn.syraIconCss + exports.cssFontIconBtn + "s-btn-" + fontIcon;
		return true;
	}
}

exports.setFontIconButton = _setFontIconButton;

exports.addFontIconText = function(text, css, fontIcon) {
	var item = document.createElement("div");
	item.className = css;
	item.textContent = text;
	item.insertBefore(exports.addFontIconFlag(css + "-flag", fontIcon), item.firstChild);
	return item;
};

exports.addFontIconFlag = function(css, fontIcon) {
	var flag = document.createElement("div");
	flag.syraIconCss = css;
	_setFontIconButton(flag, fontIcon);
	return flag;
};

exports.addIconButton = function(text, css, onClick, shortCutTip, fontIcon) {
	var btn = document.createElement("a");
	btn.syraIconCss = css;
	if (!(fontIcon && _setFontIconButton(btn, fontIcon))) {
		btn.className = css;
	}
	btn.setAttribute("href", "#"); //oneed to improve cancel action otherwiese add # to url
	btn.syraOnClick = onClick;
	btn.syraShortCutTip = shortCutTip;
	btn.title = text;
	btn.syraTip = 2;
	return btn;
};

exports.setButtonTitle = function(btn, text) {
	if (btn.syraTipTitle) {
		btn.syraTipTitle = text;
	} else {
		btn.title = text;
	}
	if (btn.syraTip != 2) {
		btn.textContent = text;
	}
};
exports.updateButtonIcon = function(btn, text, fontIcon) {
	var icon;
	if (btn.syraHasIcon) {
		btn.removeChild(icon = btn.firstChild);
	}
	exports.setButtonTitle(btn, text);
	if (icon) {
		btn.insertBefore(icon, btn.firstChild);
	}
	fontIcon && _setFontIconButton(icon || btn, fontIcon);
};

exports.onItemInOut = function(onEnter, event, target) {
	if (onEnter) {
		if (target.outOfDate) {
			target.setMenu(null, target.articleParent.dataset, target.outOfDate);
			target.outOfDate = false;
		}
		var showAlways;
		if (target.domItem) {
			showAlways = target.domItem.syraTitle && target.$item.$noText;
			target = target.domItem;
		}
		syra_site.dialogManager.titleTip.show(target.title, target, !showAlways); // if no show always => check tip visibility
	}
	target = target.domItem ? target.domItem : target;
	!target.syraPickerType && syra_site.dom.toggleClass(target, "s-enter", onEnter);
};



var _listMenuBinds = ["$edit", "$create", "$delete", "$details", "$toggleGraph", "$design", "$cancelSort"];

function _getIcon(field, $bind) {

	return null;
}

function _clearMenus(item) {
	var ids = item.mnPickers && Object.keys(item.mnPickers);
	for (var ii = 0, jj = ids.length; ii < jj; ii++) {
		syra_site.dom.removeChild(item.mnPickers[ids[ii]]);
	}
	var $binds = Object.keys(item.menuItems);
	for (var ii = 0, jj = $binds.length; ii < jj; ii++) {
		var $bind = $binds[ii];
		var mn, items = item.menuItems[$bind];
		for (var kk = 0, mm = items.length; kk < mm; kk++) {
			mn = items[kk];
			if (mn.detailitemId) {
				var item = syra_store.get(mn.detailitemId);
				item && item.setDataValue(item.currentValue);
			}
			(mn.layoutParent ? mn.layoutParent : item).removeItem(mn, true);
		}
	}
	exports.initialize(item);
	if (item.menusBox) {
		(item.menusBox.layoutParent ? item.menusBox.layoutParent : item).removeItem(item.menusBox, true);
		if (item.menusBag == item.menusBox) {
			delete item.menusBag;
		}
		delete item.menusBox;
	}
	item.mnPickers.menus && _togglePicker(item.mnPickers.menus, "none");
}



exports.removeMenu = function(article, $bind, removeDom) {
	if (article.menuItems) {
		var menuItems = article.menuItems[$bind];
		if (menuItems) {
			for (var ii = 0, jj = menuItems.length; ii < jj; ii++) {
				article.removeItem(menuItems[ii], removeDom);
			}
			delete article.menuItems[$bind];
		}
	}
	delete article.$menus[$bind];
	delete article.$menusPopup[$bind];
};

exports.addFieldPicker = _addFieldPicker;
exports.addPicker = _addPicker;


function _togglePicker(picker, syraDisplay) {
	if (picker && picker.syraDisplay != syraDisplay) {
		picker.style.display = picker.syraDisplay = syraDisplay;
	}
}

function _applyMenusChange(field, $deltaMenu, $isAction, metaData) {
	var binds = Object.keys($deltaMenu);
	for (var ii = 0, jj = binds.length; ii < jj; ii++) {
		var $bind = binds[ii];
		var $menu = $deltaMenu[$bind];
		var created;
		if ($menu === null) {
			exports.removeMenu(field, $bind);
		} else {
			if ($menu.$target == "embedded") {
				_embeddedHelper.addEmbeddedLink(field, $bind, $menu);
				continue;
			}
			if (metaData && metaData.$diagnosePanelId) {
				$menu.$sourceBind = $menu.$sourceBind || $bind;
				$bind = metaData.$diagnosePanelId + "-" + $bind;
			}
			var $defined = field.$menus[$bind];
			if ($defined) {
				syra_site.deltaManager.applyObjectDelta(field.page, $defined, $menu);
			} else {
				created = true;
				$menu.$isAction = $isAction;
				if (metaData && metaData.$diagnosePanelId) {
					$menu.$isPopup = false;
					field.loadNewItem(metaData.$diagnosesSlot, {
						$bind: $bind,
						$category: "link",
						$skin: "s-diagnose-link"
					});
				} else {
					$menu.$sourceBind = $menu.$sourceBind || $bind;
					$menu.$bind = field.isField ? (field.id + "-" + $bind) : $bind;
				}
				$defined = field.$menus[$bind] = $menu;
				if (field.isPopupMenuItem && !field.isPopupMenuItem($bind, $menu)) {
					$menu.$isPopup = false;
				} else {
					switch ($bind) {
						case "$first":
						case "$previous":
						case "$next":
						case "$last":
							field.isDesktopPage && _pagingCapability.addPagerLinksToPage(field, $bind);
							break;
						case "$edit":
							if (field.onAddMenuItem) {
								if (field.articleParent.$capability && field.articleParent.$capability.quickedit) {
									_addFieldPicker(field, "$quickEdit", syra_local.fieldQuickEdit);
									$menu.$title = syra_local.fieldTunnel;
									$menu.$iconValue = "$tunnel";
									$menu.$isPopup = true;
								} else {
									$menu.$title = $menu.$title || syra_local.fieldEdit;
									field.onAddMenuItem($bind, $menu);
									$menu.$isPopup = false;
								}
							} else {
								$menu.$isPopup = true;
							}
							break;
						case "$canceledit":
							if (field.onAddMenuItem) {
								$menu.$title = $menu.$title || syra_local.fieldEdit;
								field.onAddMenuItem($bind, $menu);
								$menu.$isPopup = false;
							} else {
								$menu.$isPopup = true;
							}
							break;
						case "$delete":
							if (field.onAddMenuItem && field.$useDeletePicker) {
								$menu.$title = $menu.$title || syra_local.fieldDelete;
								field.onAddMenuItem($bind, $menu);
								$menu.$isPopup = false;
							} else {
								$menu.$isPopup = true;
							}
							break;
						case "remove_separator":
						case "add_separator":
							field.onAddMenuItem && field.onAddMenuItem($bind, $menu);
							$menu.$isPopup = false;
							break;
						case "$create":
						case "$select":
							$menu.$isPopup = !(field.isList && field.onAddMenuItem($bind, $menu));
							break;
						case "$details":
							$menu.$isPopup = !field.isField || field.$isDetailLinkDisabled;
							break;
						case "$localize":
							if (field.$item.$useLocalizePicker) {
								$menu.$isHidden = true;
							}
							if (field.$isEditMode && !field.mnPickers.localize && field.$item.$useLocalizePicker) {
								_addFieldPicker(field, $bind, field.$menus.$localize.$title);
							}
							$menu.$isPopup = true;
							break;
						case "$tunnel":
							$menu.$title = syra_local.fieldTunnel;
							$menu.$isPopup = syra_site.isTabletDevice || (field.$isEditMode && field.$item.$inplace && !field.$item.$isFilterMode);
							if (!$menu.$isPopup && !field.mnPickers.$tunnel && !field.$item.$inplace) {
								_addFieldPicker(field, "$tunnel", syra_local.fieldTunnel, "ESC F9");
								if (!field.$isDisabled && field.$isReadOnly) {
									field.mnPickers.$tunnel.style.display = field.mnPickers.$tunnel.syraDisplay = "";
								}
							}
							break;
						case "$lookup":
							field.$isEditMode && !field.mnPickers.$lookup && _addFieldPicker(field, $bind, syra_local.fieldLookup, "ESC L, ESC F12");
							$menu.$isPopup = false;
							break;
						default:
							$menu.$isPopup = true;
							break;
					}
				}

			}
			if ($defined.$isPopup && field.$menusPopup) {
				if ($menu.$isHidden) {
					delete field.$menusPopup[$bind];
				} else {
					field.$menusPopup[$bind] = $defined;
				}
			}
			if (field.menuItems) {
				var bounds = field.menuItems[$bind];
				if (bounds) {
					//if create setMenu is already done
					for (var mm = 0, kk = bounds.length; mm < kk; mm++) {
						bounds[mm].setMenu($menu, field.dataset);
					}
				} else {
					if (field.menusBag && !(field.isDashBoard && $menu.$vignettes && !field.isTOCMenu($menu))) {
						field.menusBag.addMenuItem($menu);
					}
				}
			}
		}
	}
}

exports.initialize = function(item) {
	item.$menus = {};
	item.$menusPopup = {};
	item.menuItems = {};
	item.mnPickers = {};
};

exports.setMenus = function(field, metaData) {
	if (metaData) {
		if (field.variantItem) {
			field = field.variantItem;
		}
		if (field.isChildField) {
			exports.setMenus(field.record, {
				$links: metaData.$links,
				$actions: metaData.$actions
			});
			return;
		}
		if (field.renderIcon) {
			metaData.$links !== undefined && field.renderIcon(metaData);
			return;
		}
		// in order not to show menu menu picker for unit ref field        
		if (field.$item && !field.$item.$isMenusHidden && (!field.$item.$isUnitField || field.$isEditMode)) {
			var $menus = (!field.isList && metaData.$item) || metaData;
			if ($menus.$links !== undefined || $menus.$actions !== undefined) {
				if ($menus.$links === null) {
					_clearMenus(field);
				} else {
					$menus.$links && _applyMenusChange(field, $menus.$links, false, metaData);
					$menus.$actions && _applyMenusChange(field, $menus.$actions, true, metaData);
					if (field.mnPickers && !field.mnPickers.menus && !field.$item.$isFilterMode && !field.$item.$isPickerMenuHidden && field.$menusPopup && Object.keys(field.$menusPopup).length) {
						_addFieldPicker(field, "menus", syra_local.flActions, (field.isField && !field.isList) ? "ESC F4,ESC M" : null);
						if (!field.$isDisabled && field.$isReadOnly) {
							field.mnPickers.menus.style.display = field.mnPickers.menus.syraDisplay = "";
						}
					}
				}
			}
		}
	}
};

exports.showPopup = function(field, picker) {
	if (!field.menusPopup) {
		var $item = {
			$noText: true,
			$title: syra_local.fieldActions,
			$clientId: field.id + "-menu-picker",
			$skin: "s-mn-popus",
			$itemSkin: "s-mn-link",
			$category: "menus",
			$layout: {
				$items: []
			}
		};
		var article = (field.isDesktopPage || field.isRecordArticle) ? field : field.articleParent;
		if (field.menusBox && article.menuItems) {
			//clear previous popup menus 
			var items = field.menusBox.boxChildItems;
			for (var ii = 0, jj = items.length; ii < jj; ii++) {
				delete article.menuItems[items[ii].$bind];
				article.removeItem(items[ii]);
			}
			article.removeItem(field.menusBox, true, true);
		}
		field.menusBox = field.page.loadNewItem(null, $item, article);
		field.menusBox.contextField = field;
		var binds = Object.keys(field.$menusPopup);
		//setOrder
		var newBinds = [];
		for (var ii = 0, jj = _listMenuBinds.length; ii < jj; ii++) {
			var found = binds.indexOf(_listMenuBinds[ii]);
			if (found >= 0) {
				binds.splice(found, 1);
				newBinds.push(_listMenuBinds[ii]);
			}
		}
		binds = newBinds.concat(binds);
		for (var ii = 0, jj = binds.length; ii < jj; ii++) {
			var $menu = field.$menusPopup[binds[ii]];
			if (!$menu.$icon) {
				if (_listMenuBinds.indexOf($menu.$sourceBind) >= 0) {
					$menu.$icon = {
						$mode: "iconText",
						$path: field.$iconPath
					};
				}
			}
			var menuItem = field.menusBox.addMenuItem($menu);
			article.$menus[$menu.$bind] = $menu;
			menuItem.setMenu($menu, article.dataset);
		}
		_disableMenuItems(field);
		picker && _showPickerOpen(picker, true);
		field.menusPopup = syra_site.dialogManager.openPopup(field, {
			content: field.menusBox,
			slot: field.menusBox.body,
			picker: picker,
			position: {
				my: "right top",
				at: "right bottom",
				of: $(picker)
			},
			onClose: function() {
				picker && _showPickerOpen(picker, false);
				field.menusPopup = null;
				// field.menusBox is not clear to avoid erro on execute 
			}
		});
	} else {
		field.menusPopup.close();
	}
};

exports.applyArticleState = function(field) {
	var pickers = field.mnPickers;
	var disabled = field.$isReadOnly || field.$isDisabled;
	if (field.isPickersDisabled != disabled) {
		if (pickers) {
			field.isPickersDisabled = disabled;
			pickers.$quickEdit && syra_site.dom.disableItem(pickers.$quickEdit, field.isPickersDisabled);
			pickers.$edit && syra_site.dom.disableItem(pickers.$edit, field.isPickersDisabled);
			if (!field.page.isFusionPage) {
				pickers.$delete && syra_site.dom.disableItem(pickers.$delete, field.isPickersDisabled);
				pickers.$create && syra_site.dom.disableItem(pickers.$create, field.isPickersDisabled);
			}
		}
		if (field.isSingleRecord) {
			var menuItems = field.menuItems;
			if (menuItems) {
				menuItems.$delete && menuItems.$delete[0] && menuItems.$delete[0].disable(field.isPickersDisabled);
			}
		}
	}
	if (pickers && (field.isPickersMenuDisabled != field.$isDisabled)) {
		field.isPickersMenuDisabled = field.$isDisabled;
		pickers.menus && syra_site.dom.disableItem(pickers.menus, field.isPickersMenuDisabled);
	}
};

function _disableItem(menuItem, isDisable) {
	menuItem && menuItem.disable(isDisable);
}

function _disableMenuItems(article) {
	var menus = article.menuItems;
	if (menus) {
		var isDisable = article.isPickersDisabled = article.$isReadOnly || article.$isDisabled;
		menus.$quickEdit && _disableItem(menus.$quickEdit[0], isDisable);
		menus.$edit && _disableItem(menus.$edit[0], isDisable);
		menus.$delete && _disableItem(menus.$delete[0], isDisable);
		menus.$create && _disableItem(menus.$create[0], isDisable);
	}
}


exports.applyFieldState = function(field) {
	if (field.mnPickers) {
		var keys = Object.keys(field.mnPickers);
		var display = (field.$isDisabled || field.$isReadOnly) ? "none" : "";
		for (var ii = 0, jj = keys.length; ii < jj; ii++) {
			switch (keys[ii]) {
				case "$tunnel":
				case "menus":
					_togglePicker(field.mnPickers[keys[ii]], field.$isDisabled ? "none" : "");
					break;
				default:
					_togglePicker(field.mnPickers[keys[ii]], display);
					break;
			}
		}
	}
	_disableMenuItems(field);
};
exports.highlight = function(field, on) {
	if (field.mnPickers) {
		var keys = Object.keys(field.mnPickers);
		for (var ii = 0, jj = keys.length; ii < jj; ii++) {
			syra_site.dom.toggleClass(field.mnPickers[keys[ii]], "s-enter", on);
		}
	}
};

exports.showMenuPicker = function(field) {
	var picker = field.mnPickers.menus;
	if (picker) {
		var show = field.$isEditMode || field.hasValue();
		_togglePicker(picker, show ? "" : "none");
		if (!picker.parentNode && field.onAddMenuPicker) {
			field.onAddMenuPicker(picker);
		}
	}
};

function _showPickerOpen(picker, open) {
	var fontIcon = _fontIcons[picker.syraPickerType];
	if (fontIcon) {
		picker.innerHTML = open ? _fontIcons[picker.syraPickerType + "_open"] : fontIcon;
	} else {
		if (picker.className && picker.className.indexOf) {
			//prevent from svg picker
			syra_site.dom.toggleClass(picker, "s-open", open);
		}
	}

}

function _addFieldPicker(field, type, title, shortCutTip) {
	var css = field.$mnPickersCss;
	if (field.$item.$css) {
		css += " " + field.$item.$css;
	}
	var picker = field.mnPickers[type] = _addPicker(type, title, css, shortCutTip);
	if (field.$isDisabled || field.$isReadOnly) {
		picker.style.display = picker.syraDisplay = "none";
	}
	field.onAddMenuPicker && field.onAddMenuPicker(picker);
	return picker;
}


function _addPicker(type, title, css, shortCutTip) {
	var picker = document.createElement("a");
	var icon = (type[0] == "$") ? type.substr(1) : type;
	css += " s-btn-" + icon;
	var fontIcon = _fontIcons[icon];
	if (fontIcon) {
		css += exports.cssFontIconBtn;
		picker.innerHTML = fontIcon;
	}
	picker.className = css;
	picker.syraOnClick = "onPickerClick";
	picker.syrainout = picker.syraIsSyraMenuHandler = true;
	picker.syraTip = 2;
	picker.syraPickerType = type;
	picker.syraShortCutTip = shortCutTip;
	picker.syraDisplay = "";
	if (title) {
		picker.title = picker.syraTitle = title;
	}
	if (syra_site.enableTestRobot) {
		picker.setAttribute("data-s-picker2", type);
	}
	return picker;
}

exports.onPickerClick = function(event, picker, field) {
	if (!field.authoringNode && !field.$isDisabled) {
		if (syra_site.currentQuickEdit &&
			syra_site.currentQuickEdit.needCloseConfirm(field.page, function() {
				exports.onPickerClick(event, picker, field);
			})) {
			return;
		}
		field.page.externalAdapter.onFieldClickPicker({
			field: field,
			pickerType: picker.syraPickerType,
			doEvent: function(options) {
				switch (picker.syraPickerType) {
					case "$quickEdit":
						field.list.builder.toggleQuickEdit(field);
						break;
					case "escape":
						shortcutInvoker.applyEscape(field, null, event);
						field.focus();
						break;
					case "menus":
						exports.showPopup(field, picker);
						break;
					case "$lookup":
						if ((field.articleParent && field.articleParent.onLookupClick) ? field.articleParent.onLookupClick(field, field.mnPickers.$lookup) : true) {
							field.focus && field.focus();
							var $lookup = field.$menus.$lookup;
							if (field.isChildFieldRecord) {
								field.articleParent.currentSelectRecords = null;
							}
							syra_site.dialogManager.openLookup(field.boxParent, {
								article: field.articleParent,
								$url: $lookup.$url,
								onSelectRecord: function(selectedRecords) {
									if (!field.$isDisabled && !field.$isReadOnly) {
										if (field.isChildFieldRecord) {
											var record = selectedRecords[Object.keys(selectedRecords)[0]];
											field.articleParent.setDataBind(record.dataset);
											field.page.notifyDataChange(field.articleParent, record.dataset);
										} else {
											var record = selectedRecords[Object.keys(selectedRecords)[0]]; //single selection
											field.setDataValue($lookup.$result ? record.dataset[$lookup.$result] : record.dataset);
											field.focus();
											field.notifyFieldChange(field.currentValue);
										}

									}
								}
							});
						}
						break;
					case "choice":
						field.onChoiceClick(null, field.mnPickers.choice);
						break;
					case "$localize":
						syra_site.localizer.onMenuClick(field, field.$menus.$localize);
						break;
					case "date":
						!field.$isReadOnly && field.onDateTimePickerClick(picker, options);
						break;
					case "time":
						!field.$isReadOnly && field.onDateTimePickerClick(picker, options);
						break;
					case "clear":
						field.onClearClick();
						break;
					default:
						syra_controller.executeMenu(field.$menus[picker.syraPickerType], field.isRecordArticle ? field : field.articleParent);
						break;

				}
			}
		});
	}
};

function _addMenuItemsToGroup(group, $menus, $isAction) {
	var $binds = Object.keys($menus);
	for (var ii = 0, jj = $binds.length; ii < jj; ii++) {
		var $bind = $binds[ii];
		group.addMenuItem({
			$bind: $bind,
			$icon: $menus[$bind].$icon,
			$isAction: $isAction
		});
	}
}

exports.addMenuItemsToGroup = function(article, $menus, groupId) {
	if (groupId) {
		var group = article.idMap[groupId];
		if (group) {
			$menus.$links && _addMenuItemsToGroup(group, $menus.$links);
			$menus.$actions && _addMenuItemsToGroup(group, $menus.$actions, true);
			exports.setMenus(article, $menus);
		}
	}
};

exports.dispose = function(parent) {
	parent.menusPopup && parent.menusPopup.close();
	parent.mnPickers = parent.menusPopup = parent.$menus = parent.$menusPopup = parent.menuItems = parent.pageSteps = null;
	exports.click.dispose();
};
