// syracuse-ui/lib/menus/menuGroup.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var SectionBlock = require('syracuse-ui/lib/article/sectionBlock').SectionBlock;

function MenuGroup() {}

exports.MenuGroup = helpers.defineClass(MenuGroup, SectionBlock, {
	getDefaultTitle: function() {
		return syra_local.box_menusTitle;
	},
	loadBox: function() {
		if (this.boxParent && this.boxParent.boxChildItems) {
			this.boxParent.boxChildItems.push(this);
		}
		this.page.registerSectionBlock(this);
		switch (this.$item.$skin) {
			case "s-page-menus":
				this.$item.$skin = "s-mn-main";
				break;
			case "s-mn-h2":
				delete this.$item.$skin;
				break;
		}
		this.isSection = true;
		if (this.$item) {
			if (!this.$item.$layout) {
				this.$item.$layout = {};
			}
			if (!this.$item.$layout.$items) {
				this.$item.$layout.$items = [];
			}
		}
		this.$isAuthoringEnabled = false;
		this.boxChildItems = [];
		if (this.isMenuGroupRoot = (!this.boxParent.isMenuGroup)) {
			this.menuGroupRoot = this;
		} else {
			var parent = this.boxParent;
			while (parent) {
				if (parent.menuGroupRoot) {
					this.menuGroupRoot = parent.menuGroupRoot;
					break;
				}
			}
		}
		this.isMenuGroup = true;

		this.$childMenus = {};
		this.$item.$name = this.$item.$name || this.$clientId; //used for authoring
		if (this.$item.$isPopupContent) {
			this.$item.$opened = false;
		}
		this.drawBox();

		if (!this.$item.$isPopupContent) {
			var preferences = this.page.userPreferences && this.page.userPreferences.data;
			if (preferences) {
				preferences = preferences.data && preferences.data.$menuBarGroup;
				if (preferences) {
					var $opened = preferences[this.id];
					if ($opened != undefined && this.$opened != $opened) {
						this.openBox(this.$opened = $opened);
					}
				}
			}
		}


		if (this.$item.$isMenusBag) {
			this.articleParent.menusBag = this;
		}
		if (this.isMenuGroupRoot) {
			syra_site.dom.toggleClass(this.body, "s-menu-group-root", true);
		} else {
			if (this.boxParent.$item.$isPopupContent) {
				syra_site.dom.toggleClass(this.body, "s-menu-group-root", true);
			}
		}
	},
	_loadMenuItem: function($menuItem, slot) {
		$menuItem.$category = "link";
		switch ($menuItem.$skin) {
			case "s-page-menus-link":
				$menuItem.$skin = "s-mn-main-link";
				break;
		}
		if (!$menuItem.$ownerId && this.$item.$ownerId) {
			$menuItem.$ownerId = this.$item.$ownerId;
		}
		if (!$menuItem.$skin) {
			if (this.$item.$isPopupContent) {
				$menuItem.$skin = "s-mn-link";
			} else {
				$menuItem.$skin = this.$item.$itemSkin || (this.$skin + "-link");
			}
		}
		if (this.$item.$itemIcon) {
			if ($menuItem.$icon === undefined) {
				$menuItem.$icon = {};
			}
			if (!$menuItem.$icon.$mode) {
				$menuItem.$icon.$mode = this.$item.$itemIcon.$mode;
			}
			if (!$menuItem.$icon.$path) {
				$menuItem.$icon.$path = this.$item.$itemIcon.$path;
			}
		}
		return this.page.loadNewItem(slot, $menuItem, this);
	},
	_renderMenuItems: function($layout, slot) {
		var cssRow, width;
		if ($layout.$layoutType == "row") {
			cssRow = "s-row-cell";
			if (!$layout.$autoSize) {
				width = (100 / ($layout.$items.length || 1)) + "%";
			}
		}
		var newItem;
		for (var ii = 0, jj = $layout.$items.length; ii < jj; ii++) {
			var $child = $layout.$items[ii];
			newItem = null;
			if ($child.$bind) {
				if (!$child.$category || $child.$category == "link") {
					newItem = this._loadMenuItem($child, slot);
				}
			} else {
				if ($child.$layout || $child.$category == "menus") {
					$child.$category = "menus";
					if ($child.$skin === undefined) {
						$child.$skin = this.$skin;
					}
					if (this.$item.$itemSkin && $child.$itemSkin === undefined) {
						$child.$itemSkin = this.$item.$itemSkin;
					}
					if (this.$item.$isBoxCollapsable && $child.$isBoxCollapsable === undefined) {
						$child.$isBoxCollapsable = this.$item.$isBoxCollapsable;
					}
					newItem = this.page.loadNewItem(slot, $child, this);
				} else
				if ($child.$layoutType) {
					newItem = document.createElement("div");
					this._renderMenuItems($child, newItem);
					slot.appendChild(newItem);
				}
			}
			if (cssRow && newItem) {
				newItem = newItem.domItem || newItem;
				syra_site.dom.toggleClass(newItem, cssRow, true);
				if (width) {
					newItem.style.width = width;
				}
			}
		}
	},
	renderLayoutContent: function() {
		this._renderMenuItems(this.$item.$layout, this.body);
	},
	onOpenButtonClick: function() {
		var self = this;
		if (self.$item) {
			if (self.page.userPreferences && self.page.menuBar && !self.page.isVignettePage) {
				if (self.page.isInMenuBar(self) && !self.$item.$isPopupContent) {
					self.page.userPreferences.getForKey("$menuBarGroup", {})[self.id] = self.$opened;
					self.page.savePageDesign(true);
				}
			}
			self.openBox(self.$opened);
			if (self.$item.$isPopupContent) {
				if (!self.menusPopup) {
					self.showBody(true);
					self.body.className = "s-mn-popup-body";
					self.menusPopup = syra_site.dialogManager.openPopup(self, {
						content: self,
						slot: self.body,
						position: {
							my: "right top",
							at: "right bottom",
							of: $(self.domTitle)
						},
						onClose: function() {
							self.domTitle && syra_site.dom.toggleClass(self.domTitle, "s-close", true);
							self.$opened = false;
							self.menusPopup = null;
						}
					});
				} else {
					self.menusPopup.close();
				}
			}
		}
	},
	addMenuItem: function($menu) {
		var $item = this.$childMenus[$menu.$bind];
		if (!$item) {
			this.$childMenus[$menu.$bind] = $item = {
				$bind: $menu.$bind,
				$isAction: $menu.$isAction,
				$icon: $menu.$icon
			};
		}
		this.childMenus = this.childMenus || {};
		return (this.childMenus[$menu.$bind] = this._loadMenuItem($item, this.body, this));
	},
	applyShortCut: function(shortcuts, event) {
		var current, listMenuItem;
		if (shortcuts.enter && this._$selectedListIndex != undefined) {
			var menuItem = syra_menus.listMenuItem(this)[this._$selectedListIndex];
			if (menuItem) {
				menuItem.domItem.click();
				return true;
			}
		}
		if (shortcuts.up || shortcuts.down) {
			current = (this._$selectedListIndex != undefined) ? this._$selectedListIndex : -1;
			listMenuItem = syra_menus.listMenuItem(this);
			if (shortcuts.down) {
				current++;
				if (current >= listMenuItem.length) {
					current = 0;
				}
			} else {
				current--;
				if (current < 0) {
					current = listMenuItem.length - 1;
				}
			}
			this._$selectedListIndex = Math.max(current, 0);
			syra_menus.selectMenuItem(this);
			event.preventDefault();
			event.stopPropagation();
			return true;
		}
		return false;
	},
	applyEscape: function(shortcuts, event) {
		if (this.contextField && this.contextField.menusPopup) {
			this.contextField.menusPopup.close();
			return true;
		};
		return false;

	}
});
