// syracuse-ui/lib/menus/menuBuilder.js
"use strict";

var _menuImgs = ["$first", "$last", "$next", "$previous"];

function _setTitle(menu, $title, $iconValue) {
	menu.$title = $title;
	delete menu.domItem.syraTitle;
	var title = menu.getTitle();
	menu.domItem.title = title;
	if (menu.$item.$icon && !menu.iconValue) {
		menu.cssTitle = null;
		if (menu.$item.$icon.$mode == "icon") {
			menu.$item.$noText = true;
			menu.iconValue = menu.domItem;
			menu.cssTitle = " s-mn-icon";
		} else {
			menu.iconValue = document.createElement("div");
			var iconCss = menu.$skin + "-icon s-mn-icon";
			if (menu.$item.$css) {
				iconCss += " " + menu.$item.$css;
			}
			menu.iconValue.className = iconCss;
			menu.cssTitle = " s-mn-mode-icontext";
		}
		menu.cssUpdated = true;
	}

	if (!menu.$item.$noText) {
		menu.domItem.textContent = title;
	}
	if (menu.iconValue) {
		if ($iconValue != undefined) {
			menu.$iconValue = $iconValue;
		}
		var name = menu.$iconValue || menu.value || menu.$item.$icon.$value;
		if (!name) {
			name = menu.$item.$sourceBind || menu.$item.$bind;
		}
		var $path;
		if (syra_site.isTabletDevice && _menuImgs.indexOf(name) >= 0) {
			$path = "/syracuse-ui/themes/tablet/sage/images/" + (menu.$item.$icon.$path || "");
		} else {
			$path = syra_site.$item.$iconPath + (menu.$item.$icon.$path || "");
		}
		name = name.replace("$", "");
		menu.iconValue.style.backgroundImage = "url('" + $path + name + ".png')";
		if (menu.iconValue != menu.domItem) {
			menu.domItem.insertBefore(menu.iconValue, menu.domItem.firstChild);
		}
	}
	if (menu.$target == "blank") {
		if (!menu.external) {
			menu.external = document.createElement("span");
			menu.external.className = menu.$skin + "-external s-mn-external";
		}
		menu.domItem.appendChild(menu.external);
	}
}


function _setVariants(menu, $variants) {
	if (!menu.childVariants) {
		menu.childVariants = {};
	}
	var $keys = Object.keys($variants);
	var $startIndex = menu.layoutParent.$item.$items.indexOf(menu.$item);
	for (var ii = 0, jj = $keys.length; ii < jj; ii++) {
		if (!menu.childVariants[$keys[ii]]) {
			var $newItem = helpers.object.clone(menu.$item, true);
			$newItem.$variantItemKey = $keys[ii];
			var $variant = $variants[$keys[ii]];
			$newItem.$title = menu.$title;
			if (!$variant.$title) {
				$variant.$title = menu.$title;
				if (menu.articleParent.$field) {
					if (menu.articleParent.$field.$item && menu.articleParent.$field.$item.$variants) {
						var $field = menu.articleParent.$field.$item.$variants[$newItem.$variantItemKey];
						if ($field && $field.$title) {
							$variant.$title += " " + $field.$title;
						}
					}
				}
			}
			(menu.childVariants[$keys[ii]] = menu.layoutParent.createChildItem($newItem, null, ++$startIndex)).$isAction = menu.$isAction;
		}
	}
	menu.showItem(false);
}

function _onMenuItemClick(menu, event) {
	var doClick = true;
	if (doClick && menu.contextField) {
		doClick = menu.contextField.onMenuClick ? menu.contextField.onMenuClick(menu, event) : true;
	}
	if (doClick && menu.boxParent) {
		doClick = menu.boxParent.onMenuClick ? menu.boxParent.onMenuClick(menu, event) : true;
		if (doClick) {
			doClick = menu.articleParent.onMenuClick ? menu.articleParent.onMenuClick(menu, event) : true;
		}
	}

	if (doClick) {
		if (menu.articleParent != menu.page) {
			doClick = menu.page.onMenuClick ? menu.page.onMenuClick(menu, event) : true;
		}
		if (doClick && (menu.onMenuClick ? menu.onMenuClick(event) : true)) {
			syra_controller.executeMenu(menu);
		}
	}
	setTimeout(function() {
		if (menu) {
			var parent = menu.boxParent;
			while (parent && !parent.disposed) {
				if (parent.menusPopup) {
					parent.menusPopup.close();
					break;
				}
				parent = parent.boxParent;
			}
		}
	}, 10);
}


exports.click = function(menu, event) {
	if (!menu.$isDisabled) {
		if (menu.page) {
			if (menu.isNoSyracuseUrl) {
				event.syraRetValue = true;
				return;
			}
			if (menu.page.checkMenuDesignValidaty(menu)) {
				var isValidated = true;
				if (menu.$bind == "$save" || menu.$bind == "$saveAs") {
					isValidated = menu.page.validateFields();
				}
				if (isValidated) {
					var options = {
						isSiteMenu: (menu.page == syra_site),
						menuItem: menu,
						doEvent: function() {
							_onMenuItemClick(menu, event);
						}
					};
					if (menu.boxParent && menu.boxParent.contextField) {
						//boxParent is menubox of contextMenu
						menu.contextField = options.field = menu.boxParent.contextField;
						if (menu.contextField && menu.contextField.$item && !menu.$subRecordKey) {
							menu.$subRecordKey = menu.contextField.$item.$bind;
						}
					}
					var page = menu.page.mainPage,
						dialog = syra_site.dialogManager.getTopDialogPage();
					if (dialog && dialog._content && !dialog._content.disposed) {
						page = dialog._content;
					}
					if (!page || page.disposed) {
						page = menu.page;
					}
					if (page && !page.disposed && page.externalAdapter) {
						page.externalAdapter.onMenuItemClick(options);
					}
				}
			}
		} else {
			_onMenuItemClick(menu, event);
		}
	}
};

function _ensureCss(menu, $style, $value) {
	if (menu._isLoaded) {
		if ($style !== undefined) {
			menu.cssStyle = (menu.$style = $style) ? (" s-cst-sty-" + $style) : null;
			menu.cssUpdated = true;
		}
		if ($value !== undefined && menu.$item.$prefixValue) {
			menu.cssValue = $value ? (" " + menu.$item.$prefixValue + $value) : null;
			menu.cssUpdated = true;
		}
		if (menu.cssUpdated) {
			var css = menu.$skin;
			if (menu.cssStyle) {
				css += menu.cssStyle;
			}
			if (menu.cssValue) {
				css += menu.cssValue;
			}
			if (menu.syraIsDisabled) {
				css += "s-disabled";
			}
			if (menu.$item.$css) {
				css += " " + menu.$item.$css;
			}
			if (menu.cssTitle) {
				css += menu.cssTitle;
			}
			menu.domItem.className = css;
			menu.cssUpdated = false;
		}
	}
}


function _isHrefEnabled(menuItem) {
	return !menuItem.$isAction && !(menuItem.$parameters && (menuItem.$parameters.$actions || menuItem.$parameters.$properties)) && !(menuItem.$method && menuItem.$method != "GET");
}

function _setMenu(menuItem, $menu, record) {
	if ($menu) {
		if ($menu.$type) {
			menuItem.$sourceType = $menu.$type;
		}
		if (!menuItem._isInit) {
			menuItem.articleParent.onMenuDataFilled && menuItem.articleParent.onMenuDataFilled(menuItem);
			menuItem.showItem(true);
			menuItem.setState(menuItem.$item);
			if (!$menu.$title && menuItem.$item.$title) {
				$menu.$title = menuItem.$item.$title; //use for pager (next prev) currently
			}
			if (menuItem.$item.$value) {
				menuItem.value = menuItem.$item.$value;
			}
			menuItem._isInit = true;
		}
		if ($menu.$isSeparator || $menu.$title == "-") {
			menuItem.domItem.className = menuItem.$skin + "-separator";
		} else {
			if ($menu.$title !== undefined || $menu.$iconValue !== undefined) {
				_setTitle(menuItem, $menu.$title, $menu.$iconValue);
			}
			if ($menu.$description !== undefined) {
				menuItem.setDescription($menu.$description);
			}
			menuItem.setState($menu);
			if ($menu.$isDisabled !== undefined) {
				menuItem.disable($menu.$isDisabled);
			}
			var props = Object.keys($menu);
			for (var ii = 0, jj = props.length; ii < jj; ii++) {
				var prop = props[ii];
				menuItem[prop] = $menu[prop];
			}
			if ((menuItem.$sourceBind || menuItem.$item.$bind) == "$lazyload") {
				if (menuItem.domItem.parentNode) {
					menuItem.domItem.parentNode.removeChild(menuItem.domItem);
				}
			}
			if ($menu.$isDisabled && $menu.$isRequested === false) {
				if (menuItem.$item.onServerRequest ? menuItem.$item.onServerRequest(menuItem) : true) {
					if (menuItem.articleParent.onAfterActionMenuExecute) {
						if (menuItem.articleParent.onAfterActionMenuExecute(menuItem, $menu) === false) {
							$menu.$isRequested = true;
							return;
						}
					}
					if ($menu.$diagnoses) {
						_showDiagnoses(menuItem, $menu);
					}
					if ($menu.$clientAgent) {
						var agent = syra_site.agents[$menu.$clientAgent.$clientId];
						if (agent) {
							agent[$menu.$clientAgent.$action](menuItem, $menu, record);
						}
					}
					if ($menu.$links && $menu.$links.$location) {
						var $url = $menu.$links.$location.$url;
						setTimeout(function() {
							syra_site.unload();
							window.open($url, "_self");
						}, 100);
					}

					// TODO
					// job tracking ui
					if ($menu.$location && $menu.$state) {
						syra_site.registerJobSyra($menu.$location, $menu.$state, menuItem.$title, $menu.$location.split(/'/)[1]);
					}


				}
				$menu.$isRequested = true; //avoid processing action for other menu bound to $menu (create menu for sample)
			}
			if ($menu.$variants) {
				if (menuItem.$item.$variantItemKey) {
					_setMenu(menuItem, $menu.$variants[menuItem.$item.$variantItemKey], record);

				} else {
					_setVariants(menuItem, $menu.$variants);
				}
			}
		}
	}
	if (!menuItem.$sourceUrl || ($menu && $menu.$url)) {
		menuItem.$sourceUrl = menuItem.$url;
		if (menuItem.$type && (menuItem.$type.indexOf('json') == -1) && (menuItem.$type != "html")) {
			if (menuItem.$sourceUrl && (menuItem.$sourceUrl.indexOf("format=") < 0)) { // TODO: improve menuItem hack
				menuItem.$sourceUrl += (menuItem.$sourceUrl.indexOf("?") >= 0 ? "&" : "?") + "format=" +
					syra_site.urlMaker.formatMenuUrl(menuItem.articleParent, {
						$url: menuItem.$type,
						$parameters: menuItem.$parameters
					}, record);
			}
		}
	}
	menuItem.$url = menuItem.$sourceUrl;

	if (menuItem.$sourceType && (menuItem.$sourceType.indexOf('json') == -1) && (menuItem.$sourceType != "html")) {
		if (_isHrefEnabled(menuItem)) {
			syra_site.history.setHref(menuItem, record, true);
		}
		menuItem.value = menuItem.$sourceType.replace("application/", "").replace("/", "-");
		if (!menuItem.$item.$icon) {
			menuItem.$item.$icon = {
				$path: "file/s-link-"
			};
		}
		if (menuItem.$title !== undefined) {
			_setTitle(menuItem, menuItem.$title);
		}
	} else {
		if (menuItem.domItem && menuItem.$url && _isHrefEnabled(menuItem)) {
			syra_site.history.setHref(menuItem, record);
		}
	}
	$menu && _ensureCss(menuItem, $menu.$style, $menu.$value);
};

exports.setMenu = _setMenu;

exports.load = function(menu, record) {
	menu.$designLevel = "field";
	menu.isMenuItem = true;
	menu.domItem = document.createElement("a");
	menu.domItem.syraOnClick = "click";
	repository[menu.domItem.syraItem = menu.domItem.syrainout = menu.id] = menu;

	menu.articleParent.setArticleId(menu.domItem);
	if (menu.$item.$skin == "s-mn-h2-link") {
		delete menu.$item.$skin;
	}

	menu.$skin = menu.$item.$skin;
	if (menu.$skin == "s-page-menus-link") {
		menu.$skin = "s-mn-main-link";
	} else {
		if (!menu.$skin) {
			menu.$skin = menu.page.isDashBoard ? "s-nav-menu-link" : "s-mn-extra-link";
		}
	}
	menu.showItem(false);
	if (!(menu.$subRecordKey = menu.$item.$subRecordKey)) {
		if (menu.boxParent && menu.boxParent.contextField) {
			menu.contextField = menu.boxParent.contextField;
			if (menu.contextField.$item) {
				menu.$subRecordKey = menu.contextField.$item.$bind;
			}
		}
	}
	if (menu.$item.$isDescriptionVisible) {
		menu._description = document.createElement("div");
		menu._description.className = menu.$item.$skinDescription || (menu.$skin + "-desc");
		menu.domItem.appendChild(menu._description);
	}

	if (menu.$item.$bind) {
		var bounds = menu.articleParent.menuItems[menu.$item.$bind];
		if (!bounds) {
			bounds = menu.articleParent.menuItems[menu.$item.$bind] = [];
		}
		bounds.push(menu);
		var $menu = menu.articleParent.$menus[menu.$item.$bind];
		$menu && _setMenu(menu, $menu, record);
	} else {
		menu.$item.$link && _setMenu(menu, menu.$item.$link);
	}
	menu._isLoaded = menu.cssUpdated = true;
	_ensureCss(menu, menu.$item.$style);
	menu.layoutSlot.appendChild(menu.domItem);
	menu.ensureLayoutMode();
};


function _showDiagnoses(menu, $delta) {
	var $menu = menu.articleParent.$menus[menu.$item.$bind];
	if ($menu) {
		if ($menu.$links) {
			var $binds;
			var $links = $menu.$links || {};
			if ($delta.$links) {
				$binds = Object.keys($delta.$links);
				for (var ii = 0, jj = $binds.length; ii < jj; ii++) {
					var $bind = $binds[ii];
					var $source = $delta.$links[$bind];
					var $target = $links[$bind] = $links[$bind] || {};
					var $props = Object.keys($source);
					for (var kk = 0, mm = $props.length; kk < mm; kk++) {
						var $prop = $props[kk];
						$target[$prop] = $source[$prop];
					}
				}
			}
			$binds = Object.keys($links);
			for (var ii = 0, jj = $binds.length; ii < jj; ii++) {
				var $bind = $binds[ii];
				var $link = $links[$bind];
				$link.$bind = $bind + "-" + menu.id;
				$link.$url = syra_site.urlMaker.formatMenuUrl(menu.articleParent, $link);
			}
			syra_site.showDiagnoses({
				$links: $links,
				$diagnoses: $delta.$diagnoses
			}, menu.articleParent, {
				menu: menu
			});
		} else {
			if ($delta.$diagnoses) {
				syra_site.showDiagnoses({
					$diagnoses: $delta.$diagnoses
				}, menu.articleParent, {
					menu: menu
				});
			}
		}
	}
}
