// syracuse-ui/lib/menus/picker.js
"use strict";
exports.menus = {
	add: function(field) {
		return syra_button.addFieldPicker({
			parent: field,
			pickerId: "menus",
			fontIcon: "menus",
			text: syra_local.flActions,
			shortCutTip: (field.isField && !field.isArrayField) ? syra_shortCuts.tip.contextMenu : null,
			click: this.onClick
		});
	},
	onClick: function() {
		syra_menus.open({
			scope: this.parent,
			picker: this
		});
	},
	show: function(field) {
		if (field.picker_menus) {
			var show = field.$isEditMode || field.hasValue();
			syra_button.hide(field.picker_menus, !show);
			if (!field.picker_menus.link.parentNode && field.onAddMenuPicker) {
				field.onAddMenuPicker(field.picker_menus);
			}
		}
	}
};

exports.$localize = {
	add: function(field, $menu) {
		if (field.$item.$useLocalizePicker) {
			$menu.$isHidden = true;
		}
		if (field.$isEditMode && !field.picker_localize && field.$item.$useLocalizePicker) {
			syra_button.addFieldPicker({
				parent: field,
				pickerId: "$localize",
				fontIcon: "localize",
				text: field.$menus.$localize.$title,
				click: this.onClick
			});
		}
		$menu.$isPopup = true;
	},
	onClick: function() {
		syra_localizer.onMenuClick({
			scope: this.parent,
			menu: this.parent.$menus.$localize
		});
	}
};

exports.$tunnel = {
	add: function(field, $menu) {
		$menu.$title = syra_local.fieldTunnel;
		$menu.$isPopup = syra_site.isTabletDevice || (field.$isEditMode && field.$item.$inplace && !field.$item.$isFilterMode);
		if (!$menu.$isPopup && !field.picker_$tunnel && !field.$item.$inplace) {
			syra_button.addFieldPicker({
				parent: field,
				pickerId: "$tunnel",
				fontIcon: "tunnel",
				text: syra_local.fieldTunnel,
				shortCutTip: syra_shortCuts.tip.openTunnel,
				click: this.onClick
			});
			if (!field.$isDisabled && field.$isReadOnly) {
				syra_button.hide(field.picker_$tunnel, false);
			}
		}
	},
	onClick: function() {
		var field = this.parent;
		syra_router.executeMenu(field.$menus.$tunnel, field.isRecordArticle ? field : field.articleParent);
	}
};

exports.$lookup = {
	add: function(field, $menu) {
		if (field.$isEditMode && !field.picker_$lookup) {
			syra_button.addFieldPicker({
				parent: field,
				pickerId: "$lookup",
				fontIcon: "lookup",
				text: syra_local.fieldLookup,
				shortCutTip: syra_shortCuts.tip.openLookup,
				click: this.onClick
			});
		}
		$menu.$isPopup = false;
	},
	onClick: function() {
		var field = this.parent;
		if ((field.articleParent && field.articleParent.onLookupClick) ? field.articleParent.onLookupClick(field) : true) {
			field.focus && field.focus();
			var $lookup = field.$menus.$lookup;
			if (field.isChildFieldRecord) {
				field.articleParent.currentSelectRecords = null;
			}
			syra_over.openModal(field.boxParent, {
				article: field.articleParent,
				$url: $lookup.$url,
				onSelectRecord: function(selectedRecords) {
					if (!field.$isDisabled && !field.$isReadOnly) {
						if (field.isChildFieldRecord) {
							var record = selectedRecords[Object.keys(selectedRecords)[0]];
							field.articleParent.setValue(record.dataset);
							syra_form.updateDelta(field.articleParent, record.dataset);
						} else {
							var record = selectedRecords[Object.keys(selectedRecords)[0]]; //single selection
							field.setValue($lookup.$result ? record.dataset[$lookup.$result] : record.dataset);
							field.focus();
							syra_form.update(field, field.currentValue);
						}
					}
				}
			});
		}
	}
};
