// syracuse-ui/lib/page/navigation/megaDetailPanel.js
"use strict";
require('syracuse-ui/lib/field/custom/simpleScroller').simplebar;

function DetailPanel(navParent) {
	this.navParent = navParent;
	this._slot = this.navParent.body = syra_dom.div("s-nav-mega-detail-panel");
	this._scrollview = syra_dom.div("s-nav-mega-detail-scrollview", this._slot);
	this._body = syra_dom.div("s-nav-mega-detail-body", this._scrollview);
	this.cols = [];
	this.addColumn();
}

DetailPanel.prototype.addColumn = function() {
	var col = syra_dom.ul("s-nav-mega-detail-col", this._body);
	this.cols.push(col);
	return col;
};

DetailPanel.prototype.activate = function() {
	this.navParent.page.megaDetailPanelActivated = this;
	var subModule = this.navParent.subModule || this.navParent;
	var colsMax = (subModule.mode == "large" && sub.page.isHighRes) ? 3 : 2;
	if (this.cols.length < colsMax) {
		var slotHeight = this._slot.clientHeight;
		if (slotHeight != this.lastSlotHeight) {
			var overflowItems = [];
			var scrollCol = this.cols[this.cols.length - 1];
			while (slotHeight < scrollCol.scrollHeight && scrollCol.childNodes.length > 1) {
				overflowItems.push(scrollCol.removeChild(scrollCol.lastChild));
			}
			if (overflowItems.length) {
				var newCol = this.addColumn();
				for (var ii = overflowItems.length - 1; ii > -1; ii--) {
					newCol.appendChild(overflowItems[ii]);
					if (this.cols.length < colsMax) {
						if (slotHeight + 1 < newCol.scrollHeight && newCol.childNodes.length > 1) {
							newCol.removeChild(newCol.lastChild);
							newCol = this.addColumn();
							ii++;
						}
					}
				}
			}
		}
	}
	this.resizeItem();
};
DetailPanel.prototype.resizeItem = function() {
	if (this.cols.length) {
		this._body.style.width = (this.cols[0].clientWidth * this.cols.length) + "px";
	}
	if (!this.$$simplebar) {
		// adjust the position of the scrollbar...
		this.$$simplebar = $(this._slot).simplebar({
			css: {
				container: this._slot.className,
				content: this._body.className,
				scrollContent: this._scrollview.className,
				scrollbar: 'simplebar-scrollbar',
				scrollbarTrack: 'simplebar-track'
			},
			wrapContent: false,
			autoHide: false
		});
	} else {
		this.$$simplebar.simplebar("recalculate");
	}
};

DetailPanel.prototype.dispose = function() {
	syra_site.disposeObject(this);
};

exports.add = function(navParent) {
	return new DetailPanel(navParent);
};
