// syracuse-ui/lib/page/navigation/megaDetailPanel.js
"use strict";
var Scrollbar = require('syracuse-ui/lib/scroll/scrollbar').Scrollbar;

function DetailPanel(navParent) {
	this.navParent = navParent;
	this.$skin = this.navParent.page.$skin;
	this._slot = this.navParent.body = syra_dom.div(this.$skin + "-detail-panel");
	this._body = syra_dom.div(this.$skin + "-detail-body", this._slot);
	this._scrollview = syra_dom.div(this.$skin + "-detail-scrollview", this._body);
	this.cols = [];
	this.addColumn();
}

DetailPanel.prototype.addColumn = function() {
	var col = syra_dom.ul(this.$skin + "-detail-col", this._scrollview);
	this.cols.push(col);
	return col;
};

DetailPanel.prototype.activate = function() {
	this.navParent.page.megaDetailPanelActivated = this;
	var parentMenuItem = this.navParent.subModule || this.navParent;
	var mode = parentMenuItem.mode || (parentMenuItem.submodule && parentMenuItem.submodule.mode);
	var colsMax = (mode == "large" && parentMenuItem.page.isHighRes) ? 3 : 2;
	if (this.cols.length < colsMax) {
		var slotHeight = this._body.clientHeight;
		if (slotHeight != this.lastSlotHeight) {
			var overflowItems = [];
			var scrollCol = this.cols[this.cols.length - 1];
			while (slotHeight < scrollCol.scrollHeight && scrollCol.childNodes.length > 1) {
				overflowItems.push(scrollCol.removeChild(scrollCol.lastChild));
			}
			if (overflowItems.length) {
				var newCol = this.addColumn();
				for (var ii = overflowItems.length - 1; ii > -1; ii--) {
					newCol.appendChild(overflowItems[ii]);
					if (this.cols.length < colsMax) {
						if (slotHeight + 1 < newCol.scrollHeight && newCol.childNodes.length > 1) {
							newCol.removeChild(newCol.lastChild);
							newCol = this.addColumn();
							ii++;
						}
					}
				}
			}
		}
	}
	this.resizeItem();
};
DetailPanel.prototype.resizeItem = function() {
	if (this.cols.length) {
		this._scrollview.style.width = (this.cols[0].clientWidth * this.cols.length) + "px";
	}
	if (!this.scrollBar) {
		// adjust the position of the scrollbar...
		this.scrollBar = new Scrollbar({
			container: this._body,
			scrollview: this._scrollview
		});
	} else {
		this.scrollBar.resize();
	}
};

DetailPanel.prototype.dispose = function() {
	this.scrollBar && this.scrollBar.dispose();
	syra_site.disposeObject(this);
};

exports.add = function(navParent) {
	return new DetailPanel(navParent);
};
