// syracuse-ui/lib/page/navigation/megaDetailPanel.js
"use strict";
var SimpleBar = require('syracuse-ui/lib/field/custom/simpleScroller').simplebar;

function DetailPanel(navParent) {
	this.navParent = navParent;
	this._slot = syra_dom.div("simplebar", this.navParent.body);
	this._slot.id = "contentPanelScrollBar";
	this.cols = [];
	this.cols.push({
		slot: syra_dom.ul("s-mega-scroll-column", this._slot)
	});
}

DetailPanel.prototype.firstColumnSlot = function() {
	return this.cols[0].slot;
};
DetailPanel.prototype.activate = function() {
	this._slot.style.height = this.navParent.page.popup.clientHeight + "px";
	this.navParent.page.popup.appendChild(this.navParent.body); //popupBody
	var maxColumnCount = (this.navParent.subModule || this.navParent).maxColumns;
	if (!this.adjusted) {
		this.adjusted = true;

		/*
		 * Remove each overflowed root element from the column and place it an array, leave at least one element in the first column
		 */
		var overflowContent = [];
		while (this.navParent.body.clientHeight < this.firstColumnSlot().scrollHeight && this.firstColumnSlot().childNodes.length > 1) {
			overflowContent.push(this.firstColumnSlot().removeChild(this.firstColumnSlot().lastChild));
		}

		if (overflowContent.length) {
			/*
			 * Start adding the array elements to the new column slot in reverse order.
			 */
			var columnNumber = 1;
			var maxItemColumns = maxColumnCount === 5 ? 3 : 2;

			this.addColumn();

			var maxScrollHeight = Math.max(this.cols[0].slot.clientHeight, this.navParent.body.clientHeight);

			for (var ii = overflowContent.length - 1; ii > -1; ii--) {
				var overflowDomItem = overflowContent[ii];
				this.cols[columnNumber].slot.appendChild(overflowDomItem);

				if (columnNumber + 1 < maxItemColumns) {
					if (maxScrollHeight + 1 < this.cols[columnNumber].slot.scrollHeight && this.cols[columnNumber].slot.childNodes.length > 1) {
						this.cols[columnNumber].slot.removeChild(this.cols[columnNumber].slot.lastChild);

						/*if (columnNumber === this.columnList.lastColumn - 1) 
                         break;
                         */
						if (columnNumber === this.lastColumn - 1)
							break;
						maxScrollHeight = Math.max(this.cols[columnNumber].slot.clientHeight, maxScrollHeight);
						++columnNumber;
						this.addColumn();
						ii++;
					}
				}
			}
		}
		// adjust the position of the scrollbar...
		//$('#contentPanelScrollBar').simplebar().css("width", this.cols.length * 240 + "px");

	}

};

DetailPanel.prototype.addColumn = function() {
	this.cols.push({
		slot: syra_dom.ul("s-mega-scroll-column", this._slot)
	});
};

DetailPanel.prototype.dispose = function() {};

exports.add = function(navParent) {
	return new DetailPanel(navParent);
};
