// syracuse-ui/lib/page/navigation/menuField.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var ReferenceField = require('syracuse-ui/lib/field/referenceField').ReferenceField;

var _$executes = ["$execute_$query", "$execute_$details", "$execute_$edit", "$execute_$cube"];

function MenuField() {}

exports.MenuField = helpers.defineClass(MenuField, ReferenceField, {
	loadBox: function() {
		this.$item.$isMenusHidden = true;
		this.$item.$skin = this.page.isAdminMode ? "s-admin-field" : "s-nav-menu";
		ReferenceField.prototype.loadBox.call(this);
		this.onReorderItem(this.articleParent);
		this.articleParent.domItem.className += " " + this.$item.$skin + "-item";
	},
	isPopupMenuItem: function($bind) {
		return $bind.indexOf("$execute") != 0;
	},
	onReorderItem: function(record) {

	},
	addValueSlot: function() {
		this._core = this._dataValue = this.domItem;
	},
	render: function() {
		this.codeMenu = this.page.addItem(this._dataValue, {
			$subRecordKey: this.$item.$bind,
			$css: this.$item.$css,
			$category: "link",
			$skin: this.$skin + "-link"
		}, this.boxParent);
		this.codeMenu.referenceField = this;
		syra_fields.setDescription(this, this.$item.$description || this.$field.$item.$description);
		if (this.currentValue) {
			this._renderCurrentValue();
		}
	},
	_renderCurrentValue: function() {
		if (this.codeMenu) {
			if (this.$menus && this.$menus.$details) {
				syra_menus.menus.applyChange(this, {
					$links: {
						$details: {
							$title: this.currentValue.$value || "",
							$description: this.currentValue.$title || ""
						}
					}
				});
			}
			var value = syra_dataset.getFieldValue(this);
			if (value && value.$links) {
				for (var ii = 0, jj = _$executes.length; ii < jj; ii++) {
					var $execute = value.$links[_$executes[ii]];
					if ($execute) {
						$execute.$title = $execute.$title || this.currentValue.$value || "";
						if (value.convergenceFunction) {
							$execute.$description = syra_local.nvpFunction + ": " + value.convergenceFunction;
						} else {
							if (value.entity) {
								$execute.$description = syra_local.nvpEntity + ": " + value.entity;
								if (value.entity != helpers.string.pluralize(value.representation || "")) {
									$execute.$description += " (" + syra_local.nvpRepresentation + ": " + value.representation + ")";
								}
							}
						}
						if (!$execute.$description) {
							$execute.$description = value.$description;
						}
						this.codeMenu.isNavigationMenuItem = true;
						this.codeMenu.setMenu($execute, this.currentValue, this.currentValue);
						this.addBookmarkButton();
						break;
					}
				}
			}

		}
	},
	addBookmarkButton: function(item) {
		if (syra_site.bookmarks) {
			if (this.codeMenu.href) {
				this.bookmarkBtn = syra_button.add({
					parent: this,
					text: syra_local.bookmarksAdd,
					css: this.page.isAdminMode ? "s-nav-admin-menu-bookmark" : "s-nav-menu-bookmark",
					iconOnly: true,
					fontIcon: "bookmark_off",
					click: syra_site.bookmarks.onNavigationMenuItemAddClick
				});
				syra_site.bookmarks.checkNavigationMenuItemBookMark(this);
			} else {
				this.codeMenu.domItem.setAttribute("href", "#");
			}
		}
	}
});
