// syracuse-ui/lib/page/navigation/menuBlock.js
"use strict";
var _megaDetailPanel = require('./megaDetailPanel');

function MenuBlock(navParent, dataset) {
	syra_item.initialize(navParent.page, this);
	this.navParent = navParent;
	this.submodule = navParent.submodule || navParent;

	this.record = dataset;
	if (this.submodule == this.navParent && this.submodule.mode == "large") {
		this.$skin = this.submodule.$skin + "-large-block";
		this.megaDetailPanel = _megaDetailPanel.add(this);
	} else {
		this.$skin = this.page.$skin + "-block";
		this.body = syra_dom.div(this.$skin + "-body");
	}
	this.domItem = this.slot = syra_dom.li(this.$skin);
	this.domItem.syrainout = this.id;
	this.title = syra_dom.text(this.$skin + "-title", dataset.title, this.slot);
	this.submodule.addItems(dataset.items, this);
	if (!this.megaDetailPanel) {
		this.slot.appendChild(this.body);
	}
}

MenuBlock.prototype.onItemInOut = function(onEnter) {
	this.megaDetailPanel && onEnter && this.activate(true);
};

MenuBlock.prototype.activate = function(on) {
	if (this.megaDetailPanel) {
		this.submodule.activate(on, this);
	}
};
MenuBlock.prototype.dispose = function() {
	this.page && this.page.disposeItem(this);
};

exports.add = function(navParent, dataset) {
	return new MenuBlock(navParent, dataset);
};
