// syracuse-ui/lib/page/navigation/menuItem.js
"use strict";
var helpers = require('syracuse-core').helpers;

function MenuItem() {}

MenuItem.prototype.load = function(dataset) {
	this.dataset = dataset;
	this.domItem = syra_dom.anchor(this.$skin);
	this.domItem.textContent = dataset.title;
	this.domItem.syraItem = this.id;
	this.domItem.syraOnClick = "click";

	var _$executes = ["$execute_$query", "$execute_$details", "$execute_$edit", "$execute_$cube"];
	if (this.dataset && this.dataset.$links) {
		this.page = this.submodule.page;
		this.articleParent = this.submodule;

		for (var ii = 0, jj = _$executes.length; ii < jj; ii++) {
			var $execute = this.dataset.$links[_$executes[ii]];
			if ($execute) {
				$execute.$title = $execute.$title;
				if (this.dataset.convergenceFunction) {
					$execute.$description = syra_local.nvpFunction + ": " + this.dataset.convergenceFunction;
				} else {
					if (this.dataset.entity) {
						$execute.$description = syra_local.nvpEntity + ": " + this.dataset.entity;
						if (this.dataset.entity != helpers.string.pluralize(this.dataset.representation || "")) {
							$execute.$description += " (" + syra_local.nvpRepresentation + ": " + this.dataset.representation + ")";
						}
					}
				}
				if (!$execute.$description) {
					$execute.$description = this.dataset.$description;
				}

				this.$url = $execute.$url;
				syra_url.history.setHref(this);

				this.codeMenu = {
					href: this.href,
					$url: this.$url,
					description: this.description,
					title: this.title
				};

				this.singleField = {
					$variantItemKey: "menuItem",
					variantItem: this
				};

				this.singleField.$variantItemKey = "menuItem";
				this.isMenuItem = true;
				this.$item = {};
				break;
			}
		}
	}

	this.slot.appendChild(this.domItem);
};

MenuItem.prototype.click = function(event) {
	var self = this;
	syra_menus.click.fire({
		menu: this,
		event: event

	}, event);
	setTimeout(function() {
		self.page.togglePopup();
	}, 200);
};
MenuItem.prototype.dispose = function() {
	syra_item.unregister(this);
	syra_site.disposeObject(this);
};


exports.add = function(navParent) {
	var menu = syra_item.initialize(navParent.page, new MenuItem());
	menu.navParent = this;
	menu.submodule = navParent.submodule || navParent;
	return menu;
};
