// syracuse-ui/lib/page/navigation/modeMap.js
"use strict";
var Scrollbar = require('syracuse-ui/lib/scroll/scrollbar').Scrollbar;

function _clickExpand(event) {
	var block = this.parent;
	block.dataset.isExpanded = !(block.dataset.isExpanded !== false);
	if (event.shiftKey) {
		var children = block.navParent.children;
		for (var ii = 0, jj = children.length; ii < jj; ii++) {
			var child = children[ii];
			if (!child.isMenuItem) {
				child.dataset.isExpanded = block.dataset.isExpanded;
				_expand(child);
			}
		}
	} else {
		_expand(block);
	}
	block.page.modulesScrollBar.resize();
	block.page.contentScrollBar.resize();
}

function _expand(block) {
	if (block.dataset.isExpanded) {
		syra_dom.hide(block.body, false);
		syra_button.setText(block.titleBtn, undefined, "node_expanded");
		block.titleBtn.link.syraTipTitle = syra_local.nvpCollapseSubModule;
	} else {
		syra_dom.hide(block.body, true);
		syra_button.setText(block.titleBtn, undefined, "node_collapsed");
		block.titleBtn.link.syraTipTitle = syra_local.nvpExpandSubModule;
	}
}

function _listenContentScroll(page, on) {
	if (on) {
		if (!page.modulesContent_onScroll) {
			page.modulesContent.addEventListener("scroll", page.modulesContent_onScroll = function(event) {
				for (var ii = 0, jj = page.children.length - 1; ii <= jj; ii++) {
					if (page.children[ii].scrollTopSelection > this.scrollTop) {
						var module = page.children[Math.max((ii == jj) ? ii : --ii, 0)];
						if (page.lastSelected != module) {
							module.scrolling = true;
							module.activate(true);
							module.scrolling = false;
						}
						break;
					}
				}
			});
		}
	} else {
		page.modulesContent && page.modulesContent_onScroll && page.modulesContent.removeEventListener("scroll", page.modulesContent_onScroll);
		delete page.modulesContent_onScroll;
	}
};

exports.page = {
	load: function(page) {
		page.$skin = "s-nav-map";
		page.$prototype = syra_site.megaMenus.$prototype;
		page.dataset = syra_site.megaMenus.dataset;

		page.layoutSlot = syra_dom.div();
		page.domItem = syra_dom.div("", page.layoutSlot);
		page.body = syra_dom.div("", page.domItem);
		page.mapBody = syra_dom.div(page.$skin + "-body", page.body);
		page.modulesBarSlot = syra_dom.div(page.$skin + "-modules-slot", page.mapBody);
		page.modulesBar = syra_dom.ul(page.$skin + "-modules", page.modulesBarSlot);
		page.modulesContentSlot = syra_dom.div(page.$skin + "-modules-content-slot", page.mapBody);
		page.modulesScrollBar = new Scrollbar({
			container: page.modulesBarSlot,
			scrollview: page.modulesBar
		});
		page.modulesContent = syra_dom.ul(page.$skin + "-modules-content", page.modulesContentSlot);
		page.contentScrollBar = new Scrollbar({
			container: page.modulesContentSlot,
			scrollview: page.modulesContent
		});
		page.load();
		page.overPanel = syra_over.openModal(syra_site, {
			page: page,
			resize: function(dlg) {
				this.page.modulesBarSlot.style.height = this.page.size.maxBarHeight + "px";
				this.page.modulesContentSlot.style.height = this.page.size.maxBarHeight + "px";

				var maxContentWidth = this.page.body.clientWidth - this.page.modulesBarSlot.clientWidth - 1;
				this.page.modulesContentSlot.style.maxWidth = maxContentWidth + "px";

				var height = this.page.modulesContentSlot.clientHeight;
				this.page.scrollTopSelection = height / 2;
				dlg.center();
				for (var ii = 0, jj = this.page.children.length; ii < jj; ii++) {
					var module = this.page.children[ii];
					var offsetTop = module.content.offsetTop;
					if (offsetTop < height) {
						module.scrollTopSelection = offsetTop;
					} else {
						module.scrollTopSelection = offsetTop - this.page.scrollTopSelection;
					}
				}
				this.page.contentScrollBar.resize();
				this.page.modulesScrollBar.resize();
			}
		});
		page.onNavMenuClick = function() {
			page.overPanel && page.overPanel.close();
		};
		_listenContentScroll(page, true);
		if (page.children.length) {
			page.children[0].activate(true);
		}
		page.isMapLoaded = true;
	},
	resizeItem: function(page) {
		if (page.isMapLoaded && page.overPanel) {
			delete page.size;
			page.getPageSize();
			page.overPanel.resize();
			page.modulesScrollBar.resize();
			page.contentScrollBar.resize();
		}
	},
	dispose: function(page) {
		page.modulesScrollBar && page.modulesScrollBar.dispose();
		page.contentScrollBar && page.contentScrollBar.dispose();
		_listenContentScroll(page, false);
	}
};

exports.module = {
	create: function(module) {
		var code = module.dataset.code;
		if (code) {
			syra_dom.img(module.$skin + "-icon", syra_config.$iconPath + "modules/s_module_" + code.toLowerCase() + ".png", module.slot);
		}
		syra_dom.span(module.$skin + "-title", module.slot).textContent = module.dataset.title;
		this._showSubModules(module);
	},
	activate: function(on, module) {
		if (on) {
			if (!module.scrolling) {
				module.content.parentNode.scrollTop = module.content.offsetTop;
			} else {
				module.page.scrollToItem(module.slot, module.page.modulesBar, true);
			}
		} else {

		}
	},
	_showSubModules: function(module) {
		var page = module.page;
		module.content = syra_dom.li(page.$skin + "-module-content");
		module.titleBtn = syra_button.add({
			parent: module,
			slot: module.content,
			text: module.dataset.title,
			title: syra_local.nvpCollapseSubModule,
			css: module.$skin + "-content-title",
			fontIcon: "node_expanded",
			click: _clickExpand
		});
		module.body = syra_dom.ul(page.$skin + "-submodules", module.content);
		module.addSubModules();
		page.modulesContent.appendChild(module.content);
	}
};

exports.subModule = {
	create: function(sub) {
		sub.titleBtn = syra_button.add({
			parent: sub,
			slot: sub.slot,
			text: sub.dataset.title,
			title: syra_local.nvpCollapseSubModule,
			css: sub.$skin + "-title",
			fontIcon: "node_expanded",
			click: _clickExpand
		});
		sub.body = syra_dom.div(sub.$skin + "-body");
		sub.addItems(sub.dataset.items);
		var nodes = sub.body.childNodes;
		if (nodes.length) {
			sub.cols = [];
			for (var ii = 0; ii < 3; ii++) {
				sub.cols.push(syra_dom.ul(sub.$skin + "-body-col"));
			}
			var ii = 0;
			while (nodes.length) {
				sub.cols[ii].appendChild(nodes[0]);
				if ((++ii) == 3) {
					ii = 0;
				}
			}
			for (var ii = 0; ii < 3; ii++) {
				sub.body.appendChild(sub.cols[ii]);
			}
		}
		sub.slot.appendChild(sub.body);
	},
	activate: function(on, block) {
		if (on) {
			block.navParent.lastSelected = block;
		} else {

		}
	}
};

exports.menuBlock = {
	create: function(block) {
		block.domItem = block.slot = syra_dom.li(block.$skin);
		block.titleBtn = syra_button.add({
			parent: block,
			slot: block.slot,
			text: block.dataset.title,
			title: syra_local.nvpCollapseSubModule,
			css: block.$skin + "-title",
			fontIcon: "node_expanded",
			click: _clickExpand
		});
		block.body = syra_dom.ul(block.$skin + "-body", block.slot);
	}
};
