// syracuse-ui/lib/page/navigation/module.js
"use strict";
var _subModule = require('./subModule');

function Module() {

}

Module.prototype.click = function() {
	this.activate(true);
};
Module.prototype.onActivated = function() {
	if (!this.body) {
		this.body = syra_dom.ul(this.page.$skin + "-submodules");
		var dataRecords = this.dataset && this.dataset.submodules;
		this.children = [];
		if (dataRecords) {
			for (var ii = 0, jj = dataRecords.length; ii < jj; ii++) {
				this.children.push(_subModule.add(this, dataRecords[ii]));
			}
		}
		if (!this.children.length) {
			syra_dom.text(this.page.$skin + +"-no-submodules", syra_local.nvpNoSubModule, this.body);
		}
	}
	this.page.togglePopup(this.body);
};
Module.prototype.activate = function(on) {
	this.page.activateItem(this, on);
};
Module.prototype.onItemInOut = function(onEnter) {
	var self = this;
	if (onEnter) {
		self.hoverTimer = setTimeout(function() {
			delete self.hoverTimer;
			self.activate(true);
		}, 500);
	} else {
		if (self.hoverTimer) {
			clearTimeout(self.hoverTimer);
			delete self.hoverTimer;
		}
	}
};
Module.prototype.dispose = function() {
	if (this.subModules) {
		for (var ii = 0, jj = this.subModules.length; ii < jj; ii++) {
			this.subModules[ii].dispose();
		}
	}
	syra_item.unregister(this);
	syra_site.disposeObject(this);
};



exports.add = function(page, parentNode, dataset) {
	var module = syra_item.initialize(page, new Module());

	//article compatiblity
	module.$prototype = page.$prototype.$properties.modules.$item;
	module.dataset = dataset;
	module.navParent = module.articleParent = module.boxParent = page;

	var slot = module.slot = syra_dom.li(page.$skin + "-module");
	slot.syrainout = module.id;
	slot.syraOnClick = "click";
	slot.textContent = dataset.title;

	if (dataset.description && dataset.description != dataset.title) {
		slot.title = dataset.description;
	}
	if (dataset.code && module.icon) {
		module.icon.style.backgroundImage = "url('" + syra_config.$iconPath + "modules/s_module_" + dataset.code.toLowerCase() + ".png')";
	}
	parentNode.appendChild(slot);
	return module;
};
