// syracuse-ui/lib/page/navigation/moduleRecord.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var RecordArticle = require("syracuse-ui/lib/field/array/recordArticle").RecordArticle;

function ModuleRecord() {}

exports.ModuleRecord = helpers.defineClass(ModuleRecord, RecordArticle, {
	initializeRecord: function(options) {
		this.$defaultSkinSection = this.$defaultSkinBlock = this.list.$item.$skinCard || "s-h3";
	},
	reorderItem: function() {
		var rowIndex = 0;
		for (var ii = 0, jj = this.builder.rows.length; ii < jj; ii++) {
			this.builder.rows[ii].count = 0;
			this.builder.rows[ii].cells = [];
		}
		var rowOffset = 0;
		for (var ii = 0, jj = this.list.records.length; ii < jj; ii++) {
			var record = this.list.records[ii];
			record.row = this.builder.rows[rowOffset];
			record.cellIndex = record.row.count++;
			record.$$item.detach().appendTo($(record.row.slot));
			record.row.cells.push(record._item);
			if (record.row.count >= record.row.max) {
				rowOffset++;
			}
		}
		var lastRow = this.builder.rows[this.builder.rows.length - 1];
		var $$lastRow = $(lastRow.slot);
		var $$children = $$lastRow.children(".s-cards-cell");
		if ($$children.length) {
			for (var ii = 0; ii < $$children.length; ii++) {
				lastRow.cells.push($($$children[0]).detach().appendTo($$lastRow)[0]);
			}
		}
	},
	remove: function() {
		if (this.$$item) {
			this.$$item.remove();
		}
	},
	applyReorderCapability: function(reorder) {
		/*if (!this._reorder) {
         this._reorder = document.createElement("div");
         this._reorder.className = this.list.$skin + "-card-reorder s-list-record-reorder";
         this.slot.insertBefore(this._reorder, this.slot.firstChild);
         }*/
	},
	drawBox: function() {
		this.dataRow = this._item = document.createElement("div");
		this._item.className = "s-nav-module";
		this._item.setAttribute("data-s-record", this.$uuid);
		this.$$item = $(this._item);
		this.slot = document.createElement("div");
		this.slot.className = "s-nav-module-title";
		this._item.appendChild(this.slot);
		this.applyReorderCapability(true);

		this._menusSlot = document.createElement("div");
		this._menusSlot.className = "s-nav-module-links";
		this._item.appendChild(this._menusSlot);

		if (this.list.selector.isRowMode) {
			this.slot.setAttribute("data-s-picker", "list-selector");
			this.slot.className += " s-list-selector-row";
			this.slot.setAttribute("data-s-record", this.$uuid);
		}

		this.cardItem = this.page.loadNewItem(this.slot, {
			$bind: "title",
			$category: "field",
			$isEditMode: this.$isEditMode,
			$inplace: true
		}, this);
		this.list._core.appendChild(this._item);
	},
	_createMenusBox: function() {
		this.page.loadNewItem(this._menusSlot, {
			$noText: true,
			$category: "link",
			$skin: "s-nav-module-link",
			$bind: "$edit",
			$icon: {
				$mode: "icon",
				$path: "list/s-list-"
			}
		}, this);
		this.page.loadNewItem(this._menusSlot, {
			$noText: true,
			$category: "link",
			$skin: "s-nav-module-link",
			$bind: "$delete",
			$icon: {
				$mode: "icon",
				$path: "list/s-list-"
			}
		}, this);
	},
	onSelectMouseEvent: function(event) {
		document.site.toggleClass(this._item, "s-record-over", event.type == "mouseenter");
	},
	highlightSelection: function(selected) {
		document.site.toggleClass(this._item, "s-list-record-selected", selected);
	},
	dispose: function() {
		this.cardItem = this._item = this.slot = this._reorder = this._menusSlot = this.dataRow = null;
		RecordArticle.prototype.dispose.call(this);
	}
});
