// syracuse-ui/lib/page/navigation/moreModules.js
"use strict";

function More() {

}

More.prototype.addButton = function() {
	var self = this;
	self._btn = syra_button.add({
		parent: syra_site,
		slot: syra_dom.li(self.$skin, self.page.modulesBar),
		text: syra_local.megaMenuMore,
		title: syra_local.megaMenuMoreTooltip,
		css: self.$skin + "-link",
		click: function() {
			if (!self._popup) {
				self.page.mega_popupDlg && self.page.mega_popupDlg.close();

				var slot = syra_dom.div(self.$skin + "-popup");
				slot.appendChild(self._body);
				self._popup = syra_over.openPopup(self.page, {
					content: {},
					slot: slot,
					autoCloseBoundary: self._body,
					picker: self._btn.link,
					isParentPopup: true,
					position: {
						my: "left top",
						at: "left bottom",
						of: self._btn.link
					},
					close: function() {
						self._popup = null;
					}
				});
			} else {
				self._popup.close();
			}
		}
	});
	self._body = syra_dom.ul(self.$skin + "-popup-body");
};

More.prototype.resizeModulesBar = function(maxW) {
	var modulesBar = this.page.modulesBar;
	var itemCss = " " + this.$skin + "-popup-item";
	if (modulesBar.scrollWidth > maxW) {
		while (modulesBar.scrollWidth > maxW) {
			var item = this._btn.slot.previousSibling;
			if (!item) {
				break;
			}
			item.className += itemCss;
			this._body.insertBefore(item, this._body.firstChild);
		}
	} else {
		while (this._body.childNodes.length) {
			var node = this._body.childNodes[0];
			node.className = node.className.replace(this.$skin + "-popup-item", "");
			modulesBar.insertBefore(node, this._btn.slot);
			if (modulesBar.scrollWidth > maxW) {
				node.className += itemCss;
				this._body.insertBefore(node, this._body.firstChild);
				break;
			}
		}
	}
	this._btn.slot.style.visibility = this._body.childNodes.length ? "visible" : "hidden";
};

More.prototype.dispose = function() {
	syra_button.remove(this._btn);
};

More.prototype.setOnTop = function(topZIndex) {
	this._popup && syra_dom.setZIndex(this._popup.overSlot, false, topZIndex + 1);
};

exports.resizeModulesBar = function(page, maxW) {
	if (!page.more) {
		page.more = new More();
		page.more.$skin = page.$skin + "-module-more";
		page.more.page = page;
		page.more.addButton();
	}
	page.more.resizeModulesBar(maxW);
};
