// syracuse-ui/lib/page/navigation/navigator.js
"use strict";
var helpers = require('syracuse-core').helpers;
var DesktopPage = require('syracuse-ui/lib/page/desktopPage').DesktopPage;
var _module = require('./module');

var _mode = {
	mega: require('./modeMega'),
	map: require('./modeMap'),
	edit: require('./edit/edit'),
	breadcrumb: require('./modeBreadcrumb')
};

function Navigator() {}

helpers.defineClass(Navigator, DesktopPage, {
	load: function(initData) {
		(this.page = this).$item = {
			$layout: {
				$items: []
			}
		};

		syra_pageBuilder.initialize(this);
		this.externalAdapter = syra_site.externalAdapter;
		syra_article.beforeDraw(this);
		this.mode.page.draw && this.mode.page.draw(this, initData);
		syra_article.endDraw(this);

		this.addChildren();
		this.areModulesLoaded = true;
	},
	addChildren: function() {
		var modules = this.dataset && this.dataset.modules;
		this.children = [];
		if (modules) {
			for (var ii = 0, jj = modules.length; ii < jj; ii++) {
				this.addChild(modules[ii]);
			}
		}
		this.ensureNoChild();
	},
	ensureNoChild: function() {
		if (this.children.length) {
			if (this.noChild) {
				syra_dom.remove(this.noChild);
				delete this.noChild;
			}

		} else {
			syra_dom.empty(this.modulesBar);
			(this.noChild = syra_dom.li(this.page.$skin + "-no-modules", this.modulesBar)).textContent = syra_local.nvpNoModule;
		}
	},
	addChild: function(record) {
		var child = _module.add(this, record);
		this.children.push(child);
		return child;
	},
	clear: function() {
		syra_dom.empty(this.mega_popup);
		syra_dom.empty(this.modulesSlot);
		this.more && this.more.dispose();
		this.more = this.lastSelected = null;
		if (this.children) {
			for (var ii = 0, jj = this.children.length; ii < jj; ii++) {
				this.disposeChild(this.children[ii]);
			}
		}
	},
	disposeChild: function(child) {
		if (child) {
			if (child.children) {
				for (var ii = 0, jj = child.children.length; ii < jj; ii++) {
					this.disposeChild(child.children[ii]);
				}
			}
			child.carousel && child.carousel.dispose();
			syra_item.unregister(child);
			syra_site.disposeObject(child);
		}
	},
	dispose: function() {
		this.clear();
		this.mode && this.mode.page.dispose(this);
		DesktopPage.prototype.dispose.call(this);
	}
});

exports.toggleMegaMenus = function(show) {
	if (syra_site.megaMenus) {
		syra_dom.hide(syra_site.megaMenus.body, !show);
		syra_site.resizeItem();
	}
};


exports.showMegaMenus = function($representation) {
	var selectedEndpointUuid = syra_site.userProfile && syra_site.userProfile.getSelectedEndpoint() && syra_site.userProfile.getSelectedEndpoint().$uuid || "";
	var selectedRoleUuid = syra_site.userProfile && syra_site.userProfile.getSelectedRole() && syra_site.userProfile.getSelectedRole().$uuid || "";

	if (syra_site.megaMenus) {
		if (selectedEndpointUuid === syra_site.megaMenus.selectedEndpointUiid && selectedRoleUuid === syra_site.megaMenus.selectedRoleUiid) {
			return;
		}
		syra_site.megaMenus.dispose();
	}

	var page = syra_site.megaMenus = new Navigator();
	syra_site.megaMenus.selectedEndpointUiid = selectedEndpointUuid;
	syra_site.megaMenus.selectedRoleUiid = selectedRoleUuid;
	page.$representation = $representation;
	(page.mode = _mode.mega).page.load(page);
	syra_site.resizeItem();
};

exports.showSiteMap = function() {
	var page = new Navigator();
	page.mode = _mode.map;
	page.$prototype = syra_site.megaMenus.$prototype;
	page.dataset = syra_site.megaMenus.dataset;
	page.load();
	if (page.children.length) {
		page.children[0].activate(true);
	}
	page.isMapLoaded = true;
	page.resizeItem();
};

exports.editSiteMap = function() {
	var page = new Navigator();
	page.mode = _mode.edit;
	return page;
};

exports.showBreadcrumbs = function(breadcrumbs, options) {
	var page = syra_site.breadcrumbs = new Navigator();
	page.breadcrumbs = breadcrumbs;
	page.breadcrumbIndex = options.menu.index;

	(page.mode = _mode.breadcrumb).page.load(page, options);

	if (options.menu.index === 0) {
		page.mode.modules.activate(true, page);
	} else {
		var menuNode = page.mode.modules.getModule(breadcrumbs, options.menu.index, syra_site.breadcrumbs.children);

		switch (options.menu.index) {
			case 1:
				page.mode.module.activate(true, menuNode);
				break;
			case 2:
				page.mode.subModule.activate(true, null, menuNode);
				break;
			default:
				page.mode.menuBlock.activate(true, menuNode);
		}
	}
};
