// syracuse-ui/lib/page/navigation/subModuleMenuBlock.js
"use strict";
var _megaDetailPanel = require('./megaDetailPanel');

function SubModuleMenuBlock(navParent) {
	syra_item.initialize(navParent.page, this);
	this.navParent = navParent;
	this.submodule = navParent.submodule || navParent;
}

SubModuleMenuBlock.prototype.load = function(dataset) {
	//test par rapport au submodule
	this.record = dataset;
	if (this.submodule == this.navParent && this.submodule.mode == "large") {
		this.isHLayout = true;
		this.$skin = this.submodule.$skin + "-large-block";
		this.megaDetailPanel = _megaDetailPanel.add(this);
	} else {
		this.$skin = this.page.$skin + "-block";
		this.body = syra_dom.div(this.$skin + "-body");
	}
	this.domItem = this.slot = syra_dom.li(this.$skin);
	this.domItem.syrainout = this.id;
	this.title = syra_dom.text(this.$skin + "-title", dataset.title, this.slot);
	this.submodule.addItems(dataset.items, this);
	if (!this.megaDetailPanel) {
		this.slot.appendChild(this.body);
	}
};
SubModuleMenuBlock.prototype.onItemInOut = function(onEnter) {
	onEnter && this.activate(true);
};
SubModuleMenuBlock.prototype.onActivated = function() {
	if (this.isHLayout) {
		this.megaDetailPanel.activate();
	}
};
SubModuleMenuBlock.prototype.activate = function(on) {
	this.isHLayout && this.page.activateItem(this, on);
};
SubModuleMenuBlock.prototype.dispose = function() {
	syra_item.unregister(this);
	syra_site.disposeObject(this);
};

exports.add = function(navParent) {
	return new SubModuleMenuBlock(navParent);
};
