// syracuse-ui/lib/authoring/tools/designerLocalization.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function DesignerLocalization() {}

exports.DesignerLocalization = helpers.defineClass(DesignerLocalization, null, {
	load: function(designer, $stepSkin) {
		this.designer = designer;
		this._designedPage = this.designer.awArticle.page;
		this.designer.$prototype.$properties.$titleChoice = {
			$type: "application/x-choice",
			$isTitleHidden: true,
			$value: {
				$type: "application/x-string",
				$enum: [{
					$value: 1,
					$title: this.designer.designerLocalize.aw_titleBlank
				}, {
					$value: 2,
					$title: this.designer.designerLocalize.aw_titleHidden
				}]
			}
		};
		this.designer.$prototype.$properties.$titleLocalization = {
			$type: "application/x-choice",
			$value: {
				$type: "application/x-string",
				$enum: []
			}
		};
		this.fillTitlesChoices();
	},
	onDesignerNotifyDataChange: function(awItem, metaData, field, value) {
		switch (field.$item.$bind) {
			case "$titleChoice":
				var isTitleLocalizationDisabled = true;
				switch (value) {
					case 0:
						metaData.$isTitleHidden = false;
						if (this.designer.$prototype.$properties.$titleLocalization.$isHidden) {
							value = this.designer.selectedItemTitle;
						} else {
							isTitleLocalizationDisabled = false;
							value = this.designer.awItem.$item.$title;
							if (!value || value == "-") {
								value = this.designer.$prototype.$properties.$titleLocalization.$value.$enum[0].$value;
							}
							this.designer.applyChange({
								$titleLocalization: value
							});
						}
						metaData.$title = value;
						break;
					case 1:
						metaData.$title = "-";
						metaData.$isTitleHidden = false;
						break;
					case 2:
						metaData.$isTitleHidden = true;
						break;
				}
				this.designer.applyChange({
					$properties: {
						$titleLocalization: {
							$isDisabled: isTitleLocalizationDisabled
						}
					}
				});
				break;
			case "$titleLocalization":
				metaData.$title = value;
				break;
			default:
				return false;
		}
		return true;
	},
	fillTitlesChoices: function() {
		var $enum = [];
		if (this._designedPage.$prototype.$localization) {
			if (this._designedPage.$isFusionPage) {
				$enum = this._getTitleEnum();
			} else {
				var $keys = Object.keys(this._designedPage.$prototype.$localization);
				for (var ii = 0, jj = $keys.length; ii < jj; ii++) {
					$enum.push({
						$value: "{" + $keys[ii] + "}",
						$title: this._designedPage.$prototype.$localization[$keys[ii]]
					});
				}
			}
		}
		this.designer.$prototype.$properties.$titleLocalization.$value.$enum = $enum;
		if ($enum.length) {
			this.designer.$prototype.$properties.$titleChoice.$value.$enum.unshift({
				$value: 0,
				$title: this.designer.designerLocalize.aw_titleLocalization
			});
		} else {
			this.designer.$prototype.$properties.$titleChoice.$value.$enum.unshift({
				$value: 0,
				$title: this.designer.designerLocalize.aw_titleVisible
			});
			this.designer.$prototype.$properties.$titleLocalization.$isHidden = true;
		}
	},
	setTitleItem: function(item) {
		var $titleChoice = (item.$item.$isTitleHidden || item.$item.$title == null) ? 2 : ((item.$item.$title == "-") ? 1 : 0);
		var $titleLocalization = {
			$isDisabled: ($titleChoice != 0)
		};
		if (item.isSection && this._designedPage.$isFusionPage && !$titleLocalization.$isDisabled) {
			$titleLocalization.$value = {
				$enum: this._getTitleEnum(item.$item.$title)
			};
		}
		this.designer.applyChange({
			$titleLocalization: item.$item.$title,
			$titleChoice: $titleChoice,
			$properties: {
				$titleLocalization: $titleLocalization
			}
		});
	},
	_getTitleEnum: function($title) {
		var $enum = [];
		var $keys = Object.keys(this._designedPage.$prototype.$localization);
		var $definedTitles = this._getDefinedTitles(this.designer.awArticle);
		for (var ii = 0, jj = $keys.length; ii < jj; ii++) {
			var $key = "{" + $keys[ii] + "}";
			if (($key[2] == "B" || $key[2] == "S") && ($keys == $title || !$definedTitles[$key])) {
				$enum.push({
					$value: $key,
					$title: this._designedPage.$prototype.$localization[$keys[ii]]
				});
			}
		}
		return $enum;
	},
	_getDefinedTitles: function(section, $titles) {
		$titles = $titles || {};
		if (section.$item.$title) {
			$titles[section.$item.$title] = section.id;
		}
		if (section.childrenSection) {
			for (var ii = 0, jj = section.childrenSection.length; ii < jj; ii++) {
				var child = section.childrenSection[ii];
				if (child.isSection && !child.isMenuGroup) {
					this._getDefinedTitles(section.childrenSection[ii], $titles);
				}
			}
		}
		return $titles;
	},
	dispose: function() {
		this.designer = this._designedPage = null;
	}
});
