// syracuse-ui/lib/authoring/tools/historyTool.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function _refreshStepLinks(hist) {
	if (hist) {
		if (hist.steps.length > 1) {
			syra_menus.setButtonTitle(hist.prevBtn, syra_local.aw_previousLayoutOn.replace("{0}", hist.currentStep).replace("{1}", hist.steps.length));
			syra_site.dom.disableItem(hist.prevBtn, hist.currentStep == 0);
			if (hist.currentStep == hist.steps.length - 1) {
				syra_menus.setButtonTitle(hist.nextBtn, syra_local.aw_nextLayout);
				syra_site.dom.disableItem(hist.nextBtn, true);
			} else {
				syra_menus.setButtonTitle(hist.nextBtn, syra_local.aw_nextLayoutOn.replace("{0}", hist.currentStep + 1).replace("{1}", hist.steps.length));
				syra_site.dom.disableItem(hist.nextBtn, false);
			}
			syra_site.dom.disableItem(hist.undoAllBtn, false);
		} else {
			if (hist.nextBtn) {
				syra_menus.setButtonTitle(hist.nextBtn, syra_local.aw_nextLayout);
				syra_site.dom.disableItem(hist.prevBtn, true);
				syra_site.dom.disableItem(hist.nextBtn, true);
				syra_site.dom.disableItem(hist.undoAllBtn, true);
			}
		}
	}
}

function _updateSteps(designer, resetUpdates) {
	var hist = designer.history;
	if (!resetUpdates && hist.isUpdated) {
		if (hist.currentStep != hist.steps.length - 1) {
			hist.steps.splice(hist.currentStep + 1);
		}
		var $article = helpers.object.clone(designer.designedArticle.$item, true);
		hist.steps.push($article);
		syra_site.mobileGateway && syra_site.mobileGateway.applyChange($article);
		hist.currentStep = hist.steps.length - 1;
	} else {
		(hist.steps = []).push(designer.$sourceItem);
		hist.currentStep = 0;
	}
	_refreshStepLinks(designer.history);
}

function _onAfterStepChange(designer, $item) {
	try {
		syra_site.uiLocker.lock();
		var $bind = designer.designedItem.$item.$bind;
		var $article = helpers.object.clone($item || designer.history.steps[designer.history.currentStep], true);
		var newList = designer.designedArticle.reloadLayout($article);
		var designer = newList ? newList.designer : designer;
		_refreshStepLinks(designer.history);
		var itemToSelect = null;
		if ($bind) {
			var bounds = designer.designedArticle.boundFields[$bind];
			if (bounds) {
				for (var ii = 0, jj = bounds.length; ii < jj; ii++) {
					var bound = bounds[ii];
					if (bound && !bound.disposed && bound.layoutParent) {
						itemToSelect = bound;
						break;
					}
				}
			}
		}
		if (designer.designedArticle.page.garbage) {
			designer.designedArticle.page.garbage.loadFreeField();
		}
		syra_site.mobileGateway && syra_site.mobileGateway.applyChange($article);
		if (designer.onEndHistoryChangeStep) {
			designer.onEndHistoryChangeStep(itemToSelect || designer.designedArticle);
		}
	} finally {
		syra_site.uiLocker.unlock();
	}
}

function _onStepClick(designer, btn) {
	var hist = designer.history;
	if (btn == hist.prevBtn) {
		if (hist.currentStep) {
			hist.currentStep--;
			_onAfterStepChange(designer);
		}
	}
	if (btn == hist.nextBtn) {
		if ((hist.currentStep + 1) < hist.steps.length) {
			hist.currentStep++;
			_onAfterStepChange(designer);
		}
	}
	if (btn == hist.undoAllBtn) {
		syra_diagnose.showBox({
			$title: syra_local.aw_updateMessageTitle,
			$message: syra_local.aw_confirmUndoAll,
			$type: "question",
			$buttons: "yesno",
			callback: function(response) {
				if (response.$clientId == "yes") {
					hist.currentStep = 0;
					_updateSteps(designer, true);
					_onAfterStepChange(designer);
				}
			}
		});
	}
}

exports.load = function(designer) {
	designer.$sourceItem = helpers.object.clone(designer.designedArticle.$item, true);
	var hist = designer.history = {
		onAfterStepChange: function($item) {
			_onAfterStepChange(designer, $item);
		},
		updateSteps: function(isUpdated) {
			this.isUpdated = isUpdated !== false;
			_updateSteps(designer);
		}
	};
	designer.onHistoryStepClick = function(event, btn) {
		_onStepClick(designer, btn);
	};
	_updateSteps(designer);
	hist.prevBtn = syra_menus.addIconButton("previousLayout", "s-design-history-btn-icon s-disabled", "onHistoryStepClick", null, "reload_back");
	hist.nextBtn = syra_menus.addIconButton("nextLayout", "s-design-history-btn-icon s-disabled", "onHistoryStepClick", null, "reload_next");
	hist.undoAllBtn = syra_menus.addTextButton(syra_local.aw_undoAllLayout, "s-design-history-btn s-disabled", "onHistoryStepClick");
	hist.prevBtn.syraItem = hist.nextBtn.syraItem = hist.undoAllBtn.syraItem = designer.id;
	if (designer.leftTop) {
		designer.leftTop.appendChild(hist.prevBtn);
		designer.leftTop.appendChild(hist.nextBtn);
		designer.leftTop.appendChild(hist.undoAllBtn);
	}
};
