// syracuse-ui/lib/authoring/tools/historyTool.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function HistoryTool() {}

exports.HistoryTool = helpers.defineClass(HistoryTool, null, {
	load: function(designer) {
		this.designer = designer;
		this.id = designer.id + "historytool";
		syra_store.add(this);
		this.$sourceItem = helpers.object.clone(this.designer.designedArticle.$item, true);
		this._notifySteps();

		this._btns = {
			previousLayout: syra_menus.addIconButton("previousLayout", "s-design-history-btn-icon s-disabled", "previousLayout", null, "reload_back"),
			nextLayout: syra_menus.addIconButton("nextLayout", "s-design-history-btn-icon s-disabled", "nextLayout", null, "reload_next"),
			undoAllLayout: syra_menus.addTextButton(syra_local.aw_undoAllLayout, "s-design-history-btn s-disabled", "undoAllLayout")
		};
		this._btns.previousLayout.syraTool = this.id;
		this._btns.nextLayout.syraTool = this.id;
		this._btns.undoAllLayout.syraTool = this.id;
		if (this.designer.leftTop) {
			this.designer.leftTop.appendChild(this._btns.previousLayout);
			this.designer.leftTop.appendChild(this._btns.nextLayout);
			this.designer.leftTop.appendChild(this._btns.undoAllLayout);
		}

	},
	getCurrentStep: function() {
		return this._steps[this._currentStep];
	},
	notifyUpdate: function(isUpdated) {
		this.isUpdated = isUpdated !== false;
		this._notifySteps();
	},
	_notifySteps: function(resetUpdates) {
		if (!resetUpdates && this.isUpdated) {
			if (this._currentStep != this._steps.length - 1) {
				this._steps.splice(this._currentStep + 1);
			}
			var $article = helpers.object.clone(this.designer.designedArticle.$item, true);
			this._steps.push($article);
			this.designer.applyChangeToMobile($article);
			this._currentStep = this._steps.length - 1;
		} else {
			(this._steps = []).push(this.$sourceItem);
			this._currentStep = 0;
		}
		this.refreshStepLinks();
	},
	copyContext: function(target) {
		target._steps = this._steps;
		target.$sourceItem = this.$sourceItem;
		target._currentStep = this._currentStep;
		target.isUpdated = this.isUpdated;
		target.refreshStepLinks();
	},

	refreshStepLinks: function() {
		if (this._btns) {
			if (this._steps.length > 1) {
				syra_menus.setButtonTitle(this._btns.previousLayout, syra_local.aw_previousLayoutOn.replace("{0}", this._currentStep).replace("{1}", this._steps.length));
				syra_site.dom.toggleClass(this._btns.previousLayout, "s-disabled", this._currentStep == 0);
				if (this._currentStep == this._steps.length - 1) {
					syra_menus.setButtonTitle(this._btns.nextLayout, syra_local.aw_nextLayout);
					syra_site.dom.toggleClass(this._btns.nextLayout, "s-disabled", true);
				} else {
					syra_menus.setButtonTitle(this._btns.nextLayout, syra_local.aw_nextLayoutOn.replace("{0}", this._currentStep + 1).replace("{1}", this._steps.length));
					syra_site.dom.toggleClass(this._btns.nextLayout, "s-disabled", false);
				}
				syra_site.dom.toggleClass(this._btns.undoAllLayout, "s-disabled", false);
			} else {
				if (this._btns.nextLayout) {
					syra_menus.setButtonTitle(this._btns.nextLayout, syra_local.aw_nextLayout);
					syra_site.dom.toggleClass(this._btns.previousLayout, "s-disabled", true);
					syra_site.dom.toggleClass(this._btns.nextLayout, "s-disabled", true);
					syra_site.dom.toggleClass(this._btns.undoAllLayout, "s-disabled", true);
				}
			}
		}
	},
	previousLayout: function() {
		if (this._currentStep) {
			this._currentStep--;
			this.onAfterStepChange();
		}
	},
	nextLayout: function() {
		if ((this._currentStep + 1) < this._steps.length) {
			this._currentStep++;
			this.onAfterStepChange();
		}
	},
	undoAllLayout: function() {
		var self = this;
		syra_diagnose.showBox({
			$title: syra_local.aw_updateMessageTitle,
			$message: syra_local.aw_confirmUndoAll,
			$type: "question",
			$buttons: "yesno",
			callback: function(response) {
				if (response.$clientId == "yes") {
					self._currentStep = 0;
					self._notifySteps(true);
					self.onAfterStepChange();
				}
			}
		});
	},
	onAfterStepChange: function($item) {
		try {
			syra_site.uiLocker.lock();
			var $bind = this.designer.designedItem.$item.$bind;
			var $article = helpers.object.clone($item || this.getCurrentStep(), true);
			var newList = this.designer.designedArticle.reloadLayout($article);
			var designer = newList ? newList.designer : this.designer;
			designer.historyTool.refreshStepLinks();
			var itemToSelect = null;
			if ($bind) {
				var bounds = designer.designedArticle.boundFields[$bind];
				if (bounds) {
					for (var ii = 0, jj = bounds.length; ii < jj; ii++) {
						var bound = bounds[ii];
						if (bound && !bound.disposed && bound.layoutParent) {
							itemToSelect = bound;
							break;
						}
					}
				}
			}
			if (designer.designedArticle.page.garbage) {
				designer.designedArticle.page.garbage.loadFreeField();
			}
			designer.applyChangeToMobile($article);
			if (designer.onEndHistoryChangeStep) {
				designer.onEndHistoryChangeStep(itemToSelect || designer.designedArticle);
			}
		} finally {
			syra_site.uiLocker.unlock();
		}
	},
	dispose: function() {
		syra_store.remove(this);
		this._btns = this.designer = this._steps = this.$sourceItem = null;
	}
});
