// syracuse-ui/lib/authoring/tools/itemTool.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var ItemDDAgent = require("syracuse-ui/lib/authoring/tools/itemDDAgent").ItemDDAgent;
var RowResizer = require("syracuse-ui/lib/authoring/tools/rowResizer").RowResizer;

function ItemTool() {}

exports.ItemTool = helpers.defineClass(ItemTool, null, {
	findItem: function(domItem) {
		var item = syra_store.findItem(domItem);
		return (item && item.isTreeNode) ? item.item : item;
	},
	load: function(designer, $stepSkin) {
		this.designer = designer;
		syra_dd.pageObservers.push(this);
	},

	isDraggable: function(target, event) {
		var item = syra_store.findItem(target);
		if (item) {
			var designer = item.designer || (item.articleParent || item).designer;
			if (designer == this.designer) {
				return (item && item.isTreeNode) ? item.item : item;
			}
		}
		return null;
	},
	onDragMove: function(target, event) {
		if (syra_dd.ddAgent && syra_dd.ddAgent.itemTool != this) {
			this.setDefaultCursor();
			return true;
		}
		if (syra_dd.dropableItem && this.cellToResize) {
			var item = this.findItem(this.cellToResize.parentNode);
			if (item) {
				syra_dd.start(this, RowResizer);
				return;
			}
			return true;
		}
		if (target.className && target.className.indexOf && (target.className.indexOf("s-aw-item") >= 0 || target.className.indexOf("s-aw-add-item") >= 0)) {
			var item = syra_store.findItem(target);
			if (item) {
				var designer = item.designer || (item.articleParent || item).designer;
				item = (item && item.isTreeNode) ? item.item : item;
				if (item != designer.designedArticle) {
					if (syra_dd.dropableItem) {
						syra_dd.start(this, ItemDDAgent);
					} else {
						syra_dd.ddAgent && syra_dd.ddAgent.onDragMouseMove(target, event, item);
					}
				}
				this.designer.uiSwitchTool.toggleOverItem(item, true);
				if (target.syraSeparatorType == "row") {
					var right = target.getBoundingClientRect()[syra_site.isDocumentRTL ? "left" : "right"];
					if (event.pageX > (right - 4) && event.pageX < (right + 4)) {
						if (!this.cellToResize) {
							this.cellToResize = target;
							syra_site.layoutSlot.style.cursor = "w-resize";
						}
						return false;
					}
				}
				this.setDefaultCursor();
				return false;
			}
			if (syra_dd.ddAgent) {
				this.setDefaultCursor();
				return false;
			}
		}
		this.setDefaultCursor();
		return true;
	},
	setDefaultCursor: function() {
		if (this.cellToResize) {
			this.cellToResize = null;
			syra_site.layoutSlot.style.cursor = "default";
		}
	},
	dispose: function() {
		this.cellToResize = null;
		if (this.designer) {
			var ii = syra_dd.pageObservers.indexOf(this);
			if (ii >= 0) {
				syra_dd.pageObservers.splice(ii, 1);
			}
			this.designer = null;
		}
	}
});
