// syracuse-ui/lib/authoring/tools/itemTool.js
"use strict";
var helpers = require('syracuse-core').helpers;
var ItemDDAgent = require("syracuse-ui/lib/authoring/tools/itemDDAgent").ItemDDAgent;
var RowResizer = require("syracuse-ui/lib/authoring/tools/rowResizer").RowResizer;

function ItemTool() {}

exports.ItemTool = helpers.defineClass(ItemTool, null, {
	load: function(designer, $stepSkin) {
		this.designer = designer;
		syra_dd.addToPageObservers(this, true);
	},

	isDraggable: function(target, event) {
		var item = this.designer.findItem(target, undefined, event);
		return (item && (item.$awAddNewItem || (item != this.designer.designedArticle.page && item.page == this.designer.designedArticle.page))) ? item : null;
	},
	onDragMove: function(target, event) {
		if (syra_dd.ddAgent && syra_dd.ddAgent.designer != this.designer) {
			this.setDefaultCursor();
			return true;
		}
		if (syra_dd.dropableItem) {
			if (this.cellToResize) {
				var item = this.designer.findItem(this.cellToResize.parentNode, undefined, event);
				if (item && !item.articleParent || (item.articleParent == this.designer.designedArticle)) {
					syra_dd.start(this, RowResizer, this.designer.designedArticle.page.scrollview);
					return;
				}
				return true;
			}
		}
		if (target.className && target.className.indexOf && (target.className.indexOf("s-aw-item") >= 0 || target.className.indexOf("s-aw-add-item") >= 0)) {
			var item = this.designer.findItem(target, undefined, event);
			if (item) {
				if ((!item.articleParent || (item.articleParent == this.designer.designedArticle))) {
					if (item != this.designer.designedArticle.page) {
						if (syra_dd.dropableItem) {
							syra_dd.start(this, ItemDDAgent, this.designer.designedArticle.page.scrollview);
						} else {
							syra_dd.ddAgent && syra_dd.ddAgent.onDragMouseMove(target, event, item);
						}
					}
					if (target.syraSeparatorType == "row") {
						var right = target.getBoundingClientRect()[syra_site.isDocumentRTL ? "left" : "right"];
						if (event.pageX > (right - 4) && event.pageX < (right + 4)) {
							if (!this.cellToResize) {
								this.cellToResize = target;
								syra_site.layoutSlot.style.cursor = "w-resize";
							}
							return false;
						}
					}
					this.setDefaultCursor();
					return false;
				}
			}
			if (syra_dd.ddAgent) {
				this.setDefaultCursor();
				return false;
			}
		}
		this.setDefaultCursor();
		return true;
	},
	setDefaultCursor: function() {
		if (this.cellToResize) {
			this.cellToResize = null;
			syra_site.layoutSlot.style.cursor = "default";
		}
	},
	dispose: function() {
		this.cellToResize = null;
		if (this.designer) {
			syra_dd.addToPageObservers(this, false);
			this.designer = null;
		}
	}
});
