// syracuse-ui/lib/authoring/tools/settingBar.js
"use strict";
var _addins = [require("syracuse-ui/lib/authoring/fieldAddins/choiceFieldAddin"), require("syracuse-ui/lib/authoring/fieldAddins/documentFieldAddin"), require("syracuse-ui/lib/authoring/fieldAddins/htmlFielddAddin")];

function _addPanel($title, $skin) {
	$skin = " s-aw-type-" + $skin;
	var panel = {
		slot: document.createElement("div"),
		head: document.createElement("div"),
		body: document.createElement("div")
	};
	panel.slot.className = "s-aw-panel" + $skin;
	panel.head.className = "s-aw-panel-head" + $skin;
	panel.head.textContent = $title;
	panel.itemTitle = syra_dom.addDiv("s-aw-panel-head-title" + $skin, panel.head);
	panel.slot.appendChild(panel.head);

	panel.body.className = "s-aw-panel-body" + $skin;
	panel.slot.appendChild(panel.body);
	return panel;
}

function _addBox(container, $title, $skin) {
	var box = {};
	$skin = " s-aw-type-" + $skin;
	var slot = box.slot = document.createElement("div");
	slot.className = "s-aw-box" + $skin;

	var head = document.createElement("div");
	head.className = "s-aw-box-head" + $skin;
	head.textContent = $title;
	slot.appendChild(head);

	box.body = syra_dom.addDiv("s-aw-box-body", slot);
	container.appendChild(slot);
	return box;
}

function _addTitleBox(designer, container, $level) {
	var box = _addBox(container, syra_local.aw_sectionAppearance, $level);
	box.titleField = designer.addItem(box.body, {
		$bind: "$title",
		$isEditMode: true,
		$isLeftSpaceHidden: true
	}, designer);
	box.titleChoiceField = designer.addItem(box.body, {
		$bind: "$titleChoice",
		$isTitleHidden: true,
		$isEditMode: true,
		$choiceLayout: "1"
	}, designer);
	box.titleLocalizationField = designer.addItem(box.body, {
		$bind: "$titleLocalization",
		$isTitleHidden: true,
		$isEditMode: true,
		$format: "$combo"
	}, designer);
	return box;
}

function _addFieldsLabelBox(designer, container, $level) {
	var box = _addBox(container, syra_local.aw_sectionFields, $level);
	designer.addItem(box.body, {
		$bind: "$fieldsTitleChoice",
		$isTitleHidden: true,
		$isEditMode: true,
		$choiceLayout: "1"
	}, designer);
	box.labelAlignmentField = designer.addItem(box.body, {
		$bind: "$labelAlignment",
		$isEditMode: true,
		$choiceLayout: "1"
	}, designer);
	return box;
}

function _addBehaviorBox(designer, container, $level) {
	var box = _addBox(container, syra_local.aw_sectionBehaviour, $level);
	box.isBoxCollapsableField = designer.addItem(box.body, {
		$bind: "$isBoxCollapsable",
		$isEditMode: true,
		$isLeftSpaceHidden: true
	}, designer);

	box.isMaximizableField = designer.addItem(box.body, {
		$bind: "$isMaximizable",
		$isEditMode: true,
		$isLeftSpaceHidden: true
	}, designer);
	return box;
}

function _addVignette(designer) {
	var $url = "/sdata/syracuse/collaboration/syracuse/landingVignetteSelects/$template/$workingCopies?representation=landingVignetteSelect.$edit";
	var designedItem = designer.designedItem;
	var layoutParent = designer.designedItem.layoutParent;
	syra_over.openModal(designer, {
		article: designer,
		$url: $url,
		$isOkHidden: true,
		onValidate: function(searchPage) {
			var sels = searchPage.selectedVignettes;
			if (sels) {
				var newSelect;
				for (var ii = 0, jj = sels.length; ii < jj; ii++) {
					var sel = sels[ii];
					var newItem = {
						$title: sel.vignette.title,
						$awAddNewItem: true,
						$designLevel: "field",
						$field: {
							$type: "application/x-vignette",
							$title: sel.vignette.title,
							$location: syra_site.clone(sel.vignette.$links[Object.keys(sel.vignette.$links)[0]])
						}
					};
					var newChild = syra_layoutUpdater.onAddNewItem(designedItem.articleParent, designedItem, newItem, layoutParent.isRow ? "right" : "top");
					if (!newSelect) {
						newSelect = newChild;
					}
				}
				designer.endArticleUpdate(newSelect, true);
				layoutParent.removeItem(designedItem, true, true);
			}
		},
		close: function(isCanceled) {
			if (isCanceled) {
				designedItem.boxParent && designer.selectItem(designedItem.boxParent, true);
				layoutParent.removeItem(designedItem, true, true);
			}
			return true;
		}
	});
}


function _getRowSection(designer, slot) {
	var box = _addBox(slot, syra_local.aw_rowLayout, "layout");
	designer.addItem(box.body, {
		$category: "field",
		$bind: "$rowAlign",
		$isEditMode: true
	}, designer);
	return box;
}

function _fillRowSection(designer, box) {
	if (!box.insertRow) {
		var row = document.createElement("div");
		row.className = "s-aw-insert-row";

		var cell = document.createElement("div");
		cell.className = "s-aw-insert-row-label";
		cell.textContent = syra_local.aw_addrow;
		row.appendChild(cell);

		cell = document.createElement("div");
		cell.className = "s-aw-insert-row-btns";
		designer.onAddRowClick = function(event, btn) {
			var designer = this.parent;
			var layout = designer.awLayout ? designer.awLayout : (designer.designedItem.layoutContent || designer.designedItem.layoutParent);
			var $curRow = {
				$layoutType: layout.$layout.$layoutType,
				$widths: layout.$layout.$widths
			};
			if (!layout.layoutParent) {
				layout = syra_layoutUpdater.ensureHasParent(layout);
			}
			var index = layout.layoutParent.items.indexOf(layout);
			var newRow = layout.layoutParent.loadChildItem(null, {
				$layoutType: $curRow.$layoutType,
				$widths: $curRow.$widths
			}, this.isBefore ? index : ++index);
			designer.endArticleUpdate(newRow, true);
		};
		syra_button.add({
			parent: designer,
			slot: cell,
			text: syra_local.aw_addRowBefore,
			css: "s-aw-insert-row-btn",
			click: designer.onAddRowClick,
			isBefore: true
		});
		syra_button.add({
			parent: designer,
			slot: cell,
			text: syra_local.aw_addRowAfter,
			css: "s-aw-insert-row-btn",
			click: designer.onAddRowClick
		});
		row.appendChild(cell);
		box.body.insertBefore(box.insertRow = row, box.body.firstChild);
	}
}

function _addArticle(designer) {
	var panel = _addPanel(syra_local.aw_page, "article");
	panel.advancedSection = designer.addItem(panel.body, {
		$bind: "$isAdvanced",
		$isEditMode: true,
		$isLeftSpaceHidden: true
	}, designer);
	panel.rowSection = _getRowSection(designer, panel.body);
	panel.layoutSection = _addBox(panel.body, syra_local.aw_template, "article");
	panel.fieldsLabelSection = _addFieldsLabelBox(designer, panel.body, "article");
	return panel;
}

function _addSection(designer) {
	var panel = _addPanel(syra_local.aw_section, "section");
	panel.rowSection = _getRowSection(designer, panel.body);
	panel.apperanceSection = _addBox(panel.body, syra_local.aw_appearance, "section");
	panel.layoutSection = _addBox(panel.body, syra_local.aw_template, "section");
	panel.titleSection = _addTitleBox(designer, panel.body, "section");
	panel.behaviorSection = _addBehaviorBox(designer, panel.body, "section");
	panel.fieldsLabelSection = _addFieldsLabelBox(designer, panel.body, "section");
	return panel;
}

function _addBlock(designer) {
	var panel = _addPanel(syra_local.aw_block, "block");
	panel.rowSection = _getRowSection(designer, panel.body);
	panel.apperanceSection = _addBox(panel.body, syra_local.aw_appearance, "block");
	panel.layoutSection = _addBox(panel.body, syra_local.aw_template, "block");
	panel.titleSection = _addTitleBox(designer, panel.body, "block");
	panel.behaviorSection = _addBehaviorBox(designer, panel.body, "block");
	panel.fieldsLabelSection = _addFieldsLabelBox(designer, panel.body, "block");
	return panel;
}

function _addField(designer) {
	var panel = _addPanel(syra_local.aw_field, "field");
	panel.advancedSection = designer.addItem(panel.body, {
		$bind: "$isAdvanced",
		$isEditMode: true,
		$isLeftSpaceHidden: true
	}, designer);
	panel.rowSection = _getRowSection(designer, panel.body);
	panel.layoutSection = _addBox(panel.body, syra_local.aw_template, "field");

	var box = _addBox(panel.body, syra_local.aw_sectionFieldLabel, "field");
	box.titleChoiceField = designer.addItem(box.body, {
		$bind: "$fieldTitleChoice",
		$isTitleHidden: true,
		$isEditMode: true,
		$choiceLayout: "1"
	}, designer);
	box.isTitleHiddenField = designer.addItem(box.body, {
		$bind: "$isTitleHidden",
		$title: syra_local.aw_isLabelHidden,
		$isEditMode: true,
		$isLeftSpaceHidden: true
	}, designer);
	box.labelAlignmentField = designer.addItem(box.body, {
		$bind: "$labelAlignment",
		$title: syra_local.aw_sectionFieldLabelPosition,
		$isEditMode: true,
		$choiceLayout: "1"
	}, designer);
	panel.fieldAddinSection = _addBox(panel.body, syra_local.aw_sectionFieldWidget, "field");
	panel.fieldAddinSection.fields = [];

	for (var ii = 0, jj = _addins.length; ii < jj; ii++) {
		_addins[ii].load(designer);
		_addins[ii].addToSection(designer, panel.fieldAddinSection);
	}
	return panel;
}


function _fillTabChoice(designer, box) {
	var value = designer.designedItem.isTabLayout ? "tabs" : "stack";
	var isLinkDisabled = false;
	var $layoutTypes = ["tabs", "stack"];
	if (!box.tabChoice) {
		box.tabChoice = {};
		box.body.textContent = syra_local.aw_showAs;
		for (var ii = 0, jj = $layoutTypes.length; ii < jj; ii++) {
			var $layoutType = $layoutTypes[ii];
			box.tabChoice[$layoutType] = syra_button.add({
				parent: designer,
				slot: box.body,
				text: syra_local["aw_section_" + $layoutType],
				css: "s-aw-tab-choice",
				iconOnly: true,
				imageName: "authoring/s-layout-" + $layoutType + ".png",
				click: designer.onTemplateClick,
				$modelType: "modelTab",
				$layoutType: $layoutType
			});
		}
	}
	for (var ii = 0, jj = $layoutTypes.length; ii < jj; ii++) {
		var link = box.tabChoice[$layoutTypes[ii]].link;
		link.className = "s-aw-tab-choice" + (value == $layoutTypes[ii] ? " s-selected" : "");
		syra_dom.disable(link, isLinkDisabled);
	}
}


function _setLabelItem(designer) {
	var item = designer.designedItem;
	var alignment = designer.designedArticle.page.$isEditMode ? "t" : "ll";
	designer.applyChange({
		$labelAlignment: alignment,
		$isTitleHidden: false,
		$fieldsTitleChoice: 1
	});
	if (item.$item) {
		if (item.$item.$isTopLabelAlignment !== undefined ||
			item.$item.$fieldsIsTopLabelAlignment !== undefined ||
			item.$item.$isRightTextLabelAlignment !== undefined ||
			item.$item.$fieldsIsRightTextLabelAlignment !== undefined) {
			if (item.$item.$isTopLabelAlignment || item.$item.$fieldsIsTopLabelAlignment) {
				alignment = "t";
			} else {
				if (item.$item.$isRightTextLabelAlignment || item.$item.$fieldsIsRightTextLabelAlignment) {
					alignment = "lr";
				} else {
					alignment = "ll";
				}
			}
		}
	}
	designer.applyChange({
		$labelAlignment: alignment,
		$isTitleHidden: item.$item.$isTitleHidden,
		$fieldsTitleChoice: item.$item.$fieldsIsTitleHidden ? 2 : ((item.$item.$fieldsIsTitleEmpty) ? 1 : 0),
		$properties: {
			$isTitleHidden: {
				$isHidden: item.isTabLayout == true
			}
		}
	});
}


function SettingBar(designer, slot) {
	this.addins = _addins;
	this.designer = designer;
	this.slot = document.createElement("div");
	this.slot.className = "s-aw-properties-slot";
	this.slot.style.width = "200px";
	this.slot.syraItem = this.designer.id;
	syra_dom.hide(this.slot, false);

	this.panels = {
		article: _addArticle(designer),
		section: _addSection(designer),
		block: _addBlock(designer),
		field: _addField(designer)
	};

	slot.appendChild(this.slot);
}

SettingBar.prototype.show = function() {
	this.designer.startChange();
	var designedItem = this.designer.designedItem;
	this.selectedPanel && syra_dom.remove(this.selectedPanel.slot);
	var panel = this.selectedPanel = this.panels[designedItem.$designLevel];
	this.designer.applyChange({
		$isAdvanced: false,
	});
	this.designer._ensureSelectedItemTitle();
	//ensure default setting
	this.designer.applyChange({
		$isAdvanced: false,
		$isMaximizable: false,
		$isBoxCollapsable: false,
		$rowAlign: "top"
	});

	this.designer.applyChange({
		$isAdvanced: designedItem.$item.$isAdvanced,
		$isMaximizable: designedItem.$item.$isMaximizable,
		$isBoxCollapsable: designedItem.$item.$isBoxCollapsable,
		$rowAlign: (this.designer.awLayout ? this.designer.awLayout.$layout.$rowAlign : "top")
	});

	if (designedItem.isField) {
		for (var ii = 0, jj = _addins.length; ii < jj; ii++) {
			_addins[ii].onShow(this.designer);
		}
	}

	this.slot.appendChild(this.selectedPanel.slot);
	this.designer.setFieldTitleItem(designedItem);
	this.designer.setLocalizeTitleItem(designedItem);
	_setLabelItem(this.designer);
	var layout = this.designer.awLayout || designedItem.layoutContent;
	this.selectedPanel.colCount = layout ? layout.getColumnsCount() : 1;

	this._addExcludeMe(designedItem.$designLevel);

	var isHidden = this.designer.awLayout == null ? false : true;
	if (panel.rowSection) {
		this.designer.addPageModelLinks(panel.rowSection, designedItem, this.selectedPanel.colCount, this.designer.awLayout);
		var sectionHidden = true;
		if (this.designer.awLayout) {
			_fillRowSection(this.designer, panel.rowSection);
			sectionHidden = false;
		}
		syra_dom.hide(panel.rowSection.slot, sectionHidden);
	}
	if (panel.apperanceSection) {
		_fillTabChoice(this.designer, panel.apperanceSection);
		syra_dom.hide(panel.apperanceSection.slot, isHidden);
	}
	if (panel.layoutSection) {
		this.designer.addPageModelLinks(panel.layoutSection, designedItem, this.selectedPanel.colCount, this.designer.awLayout);
		var sectionHidden = isHidden;
		if (designedItem.$designLevel == "field") {
			if (designedItem.layoutParent && designedItem.layoutParent.isRow) {
				sectionHidden = true;
			}
		}
		syra_dom.hide(panel.layoutSection.slot, sectionHidden);
	}
	if (panel.titleSection) {
		syra_dom.toggleClass(panel.titleSection.titleLocalizationField.domItem, "s-aw-title-localization", true);
		var choice = panel.titleSection.titleChoiceField._choices[0];
		choice.slot.appendChild(panel.titleSection.titleLocalizationField.domItem);
		syra_dom.hide(panel.titleSection.slot, isHidden);
	}
	panel.behaviorSection && syra_dom.hide(panel.behaviorSection.slot, isHidden);
	panel.fieldsLabelSection && syra_dom.hide(panel.fieldsLabelSection.slot, isHidden);

	if (panel.fieldAddinSection) {
		isHidden = true;
		if (designedItem.$authoringType) {
			var fields = panel.fieldAddinSection.fields;
			for (var ii = 0, jj = fields.length; ii < jj; ii++) {
				var field = fields[ii],
					isSectionHidden = true;
				if (field.$item.$authoringType == designedItem.$authoringType) {
					if (field.$item.$authoringEditMode !== undefined) {
						isHidden = isSectionHidden = !(field.$item.$authoringEditMode == designedItem.$isEditMode);
					} else {
						isHidden = isSectionHidden = false;
					}
				}
				field.setState({
					$isHidden: isSectionHidden
				});
			}
		}
		syra_dom.hide(panel.fieldAddinSection.slot, isHidden);
	}
	this.designer.endChange();
	if (designedItem.$field && designedItem.$field.$type == "application/x-vignette") {
		!designedItem.$field.$location && _addVignette(this.designer);
	}
};

SettingBar.prototype.resizeBar = function(height) {
	if (this.slot.parentNode) {
		this.slot.style.height = height + "px";
		if (this.selectedPanel) {
			this.selectedPanel.body.style.height = (Math.max(0, height - this.selectedPanel.body.offsetTop)) + "px";
		}
		this.designer.asideParent.domItem.style.paddingRight = this.slot.clientWidth + "px";
	}
};
SettingBar.prototype.getScrollHeight = function() {
	return this.slot.scrollHeight;
};
SettingBar.prototype._addExcludeMe = function(panelId) {
	if (panelId != "article") {
		if (!this.selectedPanel.excludeMeBtn) {
			var slot = document.createElement("div");
			slot.className = "s-aw-exclude-me-slot";
			this.selectedPanel.body.insertBefore(slot, this.selectedPanel.body.firstChild);
			this.selectedPanel.excludeMeBtn = syra_button.add({
				parent: this.designer,
				slot: slot,
				text: syra_local.aw_excludeMe,
				css: "s-aw-exclude-me",
				click: function() {
					var designer = this.parent;
					var layoutParent = designer.designedItem.layoutParent;
					var excludedFields;
					if (designer.designedItem.$designLevel == "field") {
						excludedFields = [designer.designedItem];
					} else {
						excludedFields = designer.designedItem.layoutContent.getFields();
					}
					//check manadatory
					if (excludedFields) {
						for (var ii = 0, jj = excludedFields.length; ii < jj; ii++) {
							var field = excludedFields[ii];
							if (field.$item.$bind) {
								if (designer.designedArticle.garbage) {
									designer.designedArticle.garbage.onExcludeField(field.$item.$bind, true);
								}
								syra_item.unbind(field);
							}
						}
					}
					var index = layoutParent.items.indexOf(designer.designedItem);
					layoutParent.removeItem(designer.designedItem, true);
					var newItem = layoutParent.items[index] || layoutParent.items[index - 1];
					if (newItem && newItem.isSpaceBox) {
						newItem = null;
					}
					newItem = newItem || layoutParent.boxParent;
					switch (layoutParent.$layout.$layoutType) {
						case "row":
							layoutParent.addSpaceBox(index);
							break;
						case "tabs":
							if (layoutParent.items.length == 0) {
								syra_layoutUpdater.convertToNewLayout(layoutParent, {
									$layoutType: "stack",
									$width: "100"
								});
							}
							break;
					}
					designer.endArticleUpdate(newItem, true);
				}
			});
		}
		var enable = !this.designer.designedItem.$isMandatory;
		if (enable && this.designer.designedItem.layoutContent) {
			var fields = this.designer.designedItem.layoutContent.getFields();
			for (var ii = 0, jj = fields.length; enable && ii < jj; ii++) {
				enable = !fields[ii].$isMandatory;
			}
		}
		syra_button.hide(this.selectedPanel.excludeMeBtn, !enable);
	}
};

SettingBar.prototype.dispose = function() {
	if (this.selectedPanel) {
		this.selectedPanel.fieldAddinSection = null;
	}
	syra_dom.remove(this.slot);
	syra_site.disposeObject(this);
};

exports.load = function(designer, slot) {
	return new SettingBar(designer, slot);
};
