// syracuse-ui/lib/authoring/tools/templatesTool.js
"use strict";

function _createLayout(section, $newLayout) {
	syra_layoutUpdater.clearContent(section.layoutContent);
	var prevLayout = section.layoutContent.domItem;
	section.$item.$layout = $newLayout;
	section.renderLayoutContent();
	section.layoutContent.removeSpaceBox(true);
	return prevLayout;
}

function _splitInColumns(section, children, maxCol) {
	children = children || syra_layoutUpdater.extractItems(section.layoutContent);
	var prevLayout = _createLayout(section, {
		$layoutType: "row",
		$widths: syra_layoutValidator.getDefaultWidths(maxCol),
		$items: []
	});
	var row = 0,
		colIndex = 0;
	for (var ii = 0, jj = children.length; ii < jj; ii++) {
		if (!children[ii].disposed) {
			var layout;
			switch (row) {
				case 0:
					layout = section.layoutContent;
					break;
				case 1:
					layout = syra_layoutUpdater.wrapIntack(section.layoutContent, [section.layoutContent.items[colIndex]], colIndex);
					break;
				default:
					layout = section.layoutContent.items[colIndex];
					break;
			}
			colIndex++;
			layout.loadChildItem(children[ii]);
			if (colIndex == maxCol) {
				colIndex = 0;
				row++;
			}
		}
	}
	syra_dom.removeChild(prevLayout);
}

function _applySectionModel(section, $model) {
	var children = syra_layoutUpdater.extractItems(section.layoutContent);
	var prevLayout = section.layoutContent.domItem;
	syra_layoutUpdater.clearContent(section.layoutContent);
	switch ($model) {
		case "model_rows":
			section.$item.$layout = {
				$layoutType: "stack",
				$items: []
			};
			section.renderLayoutContent();
			syra_layoutUpdater.newLoadChildItems(section.layoutContent, children);
			break;
		case "model_3cols":
			_applySectionColumnsModel(section, children, 3);
			break;
		case "model_4cols":
			_applySectionColumnsModel(section, children, 4);
			break;
		case "model_tabs":
			section.$item.$layout = {
				$layoutType: "tabs",
				$items: []
			};
			section.renderLayoutContent();
			syra_layoutUpdater.clearContent(section.layoutContent);
			for (var ii = 0, jj = children.length; ii < jj; ii++) {
				section.layoutContent.ensureSection(children[ii]);
			}
			break;
	}
	syra_dom.removeChild(prevLayout);
}

function _applySectionSmartModel(section) {
	var layoutUpdater = syra_layoutUpdater;
	var children = layoutUpdater.extractItems(section.layoutContent);
	syra_layoutUpdater.clearContent(section.layoutContent);
	var prevLayout = section.layoutContent.domItem;
	if (children.length > 0) {
		var $rows = [];
		var colIndex = 0;
		for (var ii = 0, jj = children.length; ii < jj; ii++) {
			colIndex++;
			if (children[ii].isSection) {
				var subChildren = children[ii].layoutContent.getItems();
				if (subChildren && subChildren.length > 0) {
					if (subChildren[0].$field && subChildren[0].$field.$type == "application/x-array") {
						$rows.push({
							$layoutType: colIndex > 1 ? "row" : "stack",
							$items: []
						});
						colIndex = 1;
					}
				}
			} else {
				if (children[ii].$field && children[ii].$field.$type == "application/x-array") {
					if (colIndex > 1) {
						$rows.push({
							$layoutType: "row",
							$items: []
						});
					}
					$rows.push({
						$layoutType: "stack",
						$items: []
					});
					colIndex = 0;
				}
			}
			if (colIndex == 3) {
				$rows.push({
					$layoutType: "row",
					$items: []
				});
				colIndex = 0;
			}
		}
		if (colIndex) {
			$rows.push({
				$layoutType: colIndex > 1 ? "row" : "stack",
				$items: []
			});
		}
		if ($rows.length == 1) {
			section.$item.$layout = $rows[0];
			section.renderLayoutContent();
			section.layoutContent.removeSpaceBox(true);
			layoutUpdater.newLoadChildItems(section.layoutContent, children);
		} else {
			section.$item.$layout = {
				$layoutType: "stack",
				$items: $rows
			};
			section.renderLayoutContent();
			var rowIndex = colIndex = 0;
			for (var ii = 0, jj = children.length; ii < jj; ii++) {
				colIndex++;
				if (children[ii].isSection) {
					var subChildren = children[ii].layoutContent.getItems();
					if (subChildren && subChildren.length > 0) {
						if (subChildren[0].$field && subChildren[0].$field.$type == "application/x-array") {
							rowIndex++;
							colIndex = 1;
						}
					}
				} else {
					if (children[ii].$field && children[ii].$field.$type == "application/x-array") {
						if (colIndex > 1) {
							rowIndex++;
						}
						colIndex = 3;
					}
				}
				section.layoutContent.items[rowIndex].removeSpaceBox(true);
				section.layoutContent.items[rowIndex].loadChildItem(children[ii]);
				if (colIndex == 3) {
					rowIndex++;
					colIndex = 0;
				}
			}
		}
	}

	children = section.layoutContent.getItems();
	for (var ii = 0, jj = children.length; ii < jj; ii++) {
		if (children[ii].isSection) {
			_applySectionSmartModel(children[ii]);
		}
	}
	syra_dom.removeChild(prevLayout);
}

function _applySectionColumnsModel(section, children, maxCols) {
	section.$item.$layout = {
		$layoutType: "row",
		$items: []
	};
	var width = Math.ceil(100 / maxCols);
	var widths = [];
	for (var ii = 0; ii < maxCols; ii++) {
		widths.push(width);
	}
	section.$item.$layout.$widths = widths.join(",");
	if (children.length > maxCols) {
		for (var ii = 0; ii < maxCols; ii++) {
			section.$item.$layout.$items.push({
				$layoutType: "stack",
				$items: []
			});
		}
	}
	section.renderLayoutContent();
	if (children.length > maxCols) {
		var colIndex = 0;
		for (var ii = 0, jj = children.length; ii < jj; ii++) {
			section.layoutContent.items[colIndex++].loadChildItem(children[ii]);
			if (colIndex == maxCols) {
				colIndex = 0;
			}
		}
	} else {
		syra_layoutUpdater.clearContent(section.layoutContent);
		syra_layoutUpdater.newLoadChildItems(section.layoutContent, children);
	}
}

function _addModelBtn(designer, $model, slot) {
	var $item = {
		parent: designer,
		slot: slot,
		text: syra_local["aw_model_" + $model] || syra_local.aw_model_default_title,
		css: "s-aw-template",
		btnclick: designer.onTemplateClick,
		$modelType: "model_" + $model,
		$model: $model
	};
	if ($model != "factory") {
		$item.iconOnly = $model != "factory";
		$item.imageName = "authoring/s-template-" + $model + ".png";
	}
	return syra_button.add($item);
}


function _applyPageFactoryModel(page) {
	var children = syra_layoutUpdater.extractItems(page.layoutContent);
	syra_layoutUpdater.clearContent(page.layoutContent);
	page.$item = syra_site.clone(syra_pageBuilder.ensureDefaultArticle(page, page.$prototype.$article, page.$prototype));
	if (!page.$item.$isSimplified && page.isFusionPage) {
		page.$item.$layout = syra_layoutUpdater.cleanConvergenceSetting(page.$item.$layout);
		page.$item.$isSimplified = true;
	}
	page.reloadLayout(page.$item);
}

function _applyCardListFactoryModel(list) {
	if (list.$item.$format == "grid") {
		var $cardPosition = list.$item.$cardItem.$position;
		list.applyDesignMeta({
			$cardPosition: "no"
		}, true);
		list.$item.$cardItem = list.defineDefaultCard();
		list.$item.$cardItem.$position = $cardPosition;
		if (list.$item.$cardItem.$position == "row") {
			list.builder.$rowFieldBinds = syra_pageBuilder.getDefinedFieldBinds(list.$item.$cardItem.$layout.$items);
		}
		list.applyDesignMeta({
			$cardPosition: $cardPosition
		}, true);
	} else {
		list.$item.$layout = list.defineDefaultCard().$layout;
		designedArticle.reloadLayout({
			$layout: list.defineDefaultCard().$layout
		});
	}
}

function _convertByColumns($colCount, designedItem, awLayout) {
	var layout = awLayout ? awLayout : designedItem.layoutContent;
	if (designedItem.$designLevel == "field") {
		layout = syra_layoutUpdater.wrapIntack(designedItem.layoutParent, [designedItem], designedItem.layoutParent.items.indexOf(designedItem));
		layout.removeSpaceBox(true);
		awLayout = layout;
	}
	var $widths = ["100", "50,50", "33,33,33", "25,25,25,25", "20,20,20,20,20", "17,17,16,16,17,17"][$colCount - 1];
	syra_layoutUpdater.convertToNewLayout(layout, {
		$layoutType: $widths == "100" ? "stack" : "row",
		$widths: $widths
	});
	return awLayout ? awLayout : designedItem.layoutContent;
}


function _applyArticleSmartModel(designer) {
	var article = designer.getDesignedCardItem ? designer.getDesignedCardItem() : designer.designedArticle;
	var children = syra_layoutUpdater.extractItems(article.layoutContent);
	syra_layoutUpdater.clearContent(article.layoutContent);
	var prevLayout = article.layoutContent.domItem;
	article.$item.$layout = {
		$layoutType: "stack",
		$items: []
	};
	article.renderLayoutContent();
	syra_layoutUpdater.clearContent(article.layoutContent);
	if (children.length > 0) {
		article.layoutContent.ensureSection(children[0]);
		if (children.length > 1) {
			if (children.length > 2) {
				var tabs = article.layoutContent.loadChildItem(null, {
					$layoutType: "tabs",
					$items: []
				});
				for (var ii = 1, jj = children.length - 1; ii < jj; ii++) {
					tabs.ensureSection(children[ii]);
				}
			}
			article.layoutContent.ensureSection(children[children.length - 1]);
		}
	}
	children = article.layoutContent.getItems();
	for (var ii = 0, jj = children.length; ii < jj; ii++) {
		_applySectionSmartModel(children[ii]);
	}
	syra_dom.removeChild(prevLayout);
}

var _builder = {
	2: function(designedItem, template, fields) {
		var prevLayout = _createLayout(designedItem, (template == "2-1") ? {
			$layoutType: "stack",
			$items: []
		} : {
			$layoutType: "row",
			$widths: "50,50",
			$items: []
		});
		syra_layoutUpdater.newLoadChildItems(designedItem.layoutContent, fields);
		syra_dom.removeChild(prevLayout);
	},
	3: function(designedItem, template, fields) {
		var prevLayout;
		var layoutUpdater = syra_layoutUpdater;
		switch (template) {
			case "3-1":
				prevLayout = _createLayout(designedItem, {
					$layoutType: "stack",
					$items: [{
						$layoutType: "row",
						$widths: "50,50",
						$items: []
					}]
				});
				designedItem.layoutContent.items[0].removeSpaceBox(true);
				layoutUpdater.newLoadChildItems(designedItem.layoutContent.items[0], fields.splice(0, 2));
				layoutUpdater.newLoadChildItems(designedItem.layoutContent, fields);
				break;
			case "3-2":
				prevLayout = _createLayout(designedItem, {
					$layoutType: "row",
					$widths: "50,50",
					$items: []
				});
				layoutUpdater.newLoadChildItems(designedItem.layoutContent.loadChildItem(null, {
					$layoutType: "stack",
					$items: []
				}), fields.splice(0, 2));
				layoutUpdater.newLoadChildItems(designedItem.layoutContent, fields);
				break;
			case "3-3":
				prevLayout = _createLayout(designedItem, {
					$layoutType: "row",
					$widths: "50,50",
					$items: []
				});
				layoutUpdater.newLoadChildItems(designedItem.layoutContent, fields.splice(0, 1));
				layoutUpdater.newLoadChildItems(designedItem.layoutContent.loadChildItem(null, {
					$layoutType: "stack",
					$items: []
				}), fields);
				break;
			case "3-4":
				prevLayout = _createLayout(designedItem, {
					$layoutType: "stack",
					$items: []
				});
				layoutUpdater.newLoadChildItems(designedItem.layoutContent, fields.splice(0, 1));
				var row = designedItem.layoutContent.loadChildItem(null, {
					$layoutType: "row",
					$widths: "50,50",
					$items: []
				});
				row.removeSpaceBox(true);
				layoutUpdater.newLoadChildItems(row, fields);
				break;
		}
		syra_dom.removeChild(prevLayout);
	},
	4: function(designedItem, template, fields) {
		var prevLayout;
		var layoutUpdater = syra_layoutUpdater;
		switch (template) {
			case "4-1":
				prevLayout = _createLayout(designedItem, {
					$layoutType: "stack",
					$items: [{
						$layoutType: "row",
						$widths: "50,50",
						$items: []
					}, {
						$layoutType: "row",
						$widths: "50,50",
						$items: []
					}]
				});
				var row = designedItem.layoutContent.items[0];
				row.removeSpaceBox(true);
				layoutUpdater.newLoadChildItems(row, fields.splice(0, 2));
				row = designedItem.layoutContent.items[1];
				row.removeSpaceBox(true);
				layoutUpdater.newLoadChildItems(row, fields);
				break;
			case "4-2":
				prevLayout = _createLayout(designedItem, {
					$layoutType: "stack",
					$items: []
				});
				layoutUpdater.newLoadChildItems(designedItem.layoutContent, fields.splice(0, 1));
				var row = designedItem.layoutContent.loadChildItem(null, {
					$layoutType: "row",
					$widths: "33,33,33",
					$items: []
				});
				row.removeSpaceBox(true);
				layoutUpdater.newLoadChildItems(row, fields);
				break;
			case "4-3":
				prevLayout = _createLayout(designedItem, {
					$layoutType: "stack",
					$items: [{
						$layoutType: "row",
						$widths: "33,33,33",
						$items: []
					}]
				});
				var row = designedItem.layoutContent.items[0];
				row.removeSpaceBox(true);
				layoutUpdater.newLoadChildItems(row, fields.splice(0, 3));
				layoutUpdater.newLoadChildItems(designedItem.layoutContent, fields);
				break;

		}
		syra_dom.removeChild(prevLayout);
	},
	5: function(designedItem, template, fields) {
		var prevLayout;
		var layoutUpdater = syra_layoutUpdater;
		switch (template) {
			case "5-1":
				prevLayout = _createLayout(designedItem, {
					$layoutType: "row",
					$widths: "50,50",
					$items: [{
						$layoutType: "stack",
						$items: []
					}, {
						$layoutType: "stack",
						$items: []
					}]
				});
				layoutUpdater.newLoadChildItems(designedItem.layoutContent.items[0], fields.splice(0, 2));
				layoutUpdater.newLoadChildItems(designedItem.layoutContent.items[1], fields);
				break;
			case "5-2":
				prevLayout = _createLayout(designedItem, {
					$layoutType: "stack",
					$items: [{
						$layoutType: "row",
						$widths: "33,33,33",
						$items: []
					}, {
						$layoutType: "row",
						$widths: "50,50",
						$items: []
					}]
				});
				var row = designedItem.layoutContent.items[0];
				row.removeSpaceBox(true);
				layoutUpdater.newLoadChildItems(row, fields.splice(0, 3));
				row = designedItem.layoutContent.items[1];
				row.removeSpaceBox(true);
				layoutUpdater.newLoadChildItems(row, fields);
				break;
			case "5-3":
				prevLayout = _createLayout(designedItem, {
					$layoutType: "stack",
					$items: []
				});
				layoutUpdater.newLoadChildItems(designedItem.layoutContent, fields.splice(0, 1));
				var row = designedItem.layoutContent.loadChildItem(null, {
					$layoutType: "row",
					$widths: "50,50",
					$items: []
				});
				row.removeSpaceBox(true);
				layoutUpdater.newLoadChildItems(row, fields.splice(0, 2));
				var row = designedItem.layoutContent.loadChildItem(null, {
					$layoutType: "row",
					$widths: "50,50",
					$items: []
				});
				row.removeSpaceBox(true);
				layoutUpdater.newLoadChildItems(row, fields);
				break;

		}
		syra_dom.removeChild(prevLayout);
	},
	6: function(designedItem, template, fields) {
		var prevLayout;
		switch (template) {
			case "6-1":
				prevLayout = _createLayout(designedItem, {
					$layoutType: "stack",
					$items: [{
						$layoutType: "row",
						$widths: "50,50",
						$items: []
					}, {
						$layoutType: "row",
						$widths: "50,50",
						$items: []
					}, {
						$layoutType: "row",
						$widths: "50,50",
						$items: []
					}]
				});
				var row = designedItem.layoutContent.items[0];
				row.removeSpaceBox(true);
				syra_layoutUpdater.newLoadChildItems(row, fields.splice(0, 2));
				row = designedItem.layoutContent.items[1];
				row.removeSpaceBox(true);
				syra_layoutUpdater.newLoadChildItems(row, fields.splice(0, 2));
				row = designedItem.layoutContent.items[2];
				row.removeSpaceBox(true);
				syra_layoutUpdater.newLoadChildItems(row, fields.splice(0, 2));
				break;
			case "6-2":
				prevLayout = _createLayout(designedItem, {
					$layoutType: "stack",
					$items: [{
						$layoutType: "row",
						$widths: "33,33,33",
						$items: []
					}, {
						$layoutType: "row",
						$widths: "33,33,33",
						$items: []
					}]
				});
				var row = designedItem.layoutContent.items[0];
				row.removeSpaceBox(true);
				syra_layoutUpdater.newLoadChildItems(row, fields.splice(0, 3));
				row = designedItem.layoutContent.items[1];
				row.removeSpaceBox(true);
				syra_layoutUpdater.newLoadChildItems(row, fields.splice(0, 3));
				break;

		}
		syra_dom.removeChild(prevLayout);
	},
	rows: function(designedItem, template, fields) {
		syra_dom.removeChild(prevLayout);
	},
	"3cols": function(designedItem, template, fields) {
		_splitInColumns(designedItem, fields, 3);
	},
	"4cols": function(designedItem, template, fields) {
		_splitInColumns(designedItem, fields, 4);
	},
	mixt: function(section, template, children) {
		children = children || syra_layoutUpdater.extractItems(section.layoutContent);
		var prevLayout = section.layoutContent.domItem;
		syra_layoutUpdater.clearContent(section.layoutContent);
		section.$item.$layout = {
			$layoutType: "row",
			$widths: "50,50",
			$items: []
		};
		section.renderLayoutContent();
		syra_layoutUpdater.clearContent(section.layoutContent);
		if (children.length) {
			section.layoutContent.loadChildItem(children[0]);
			if (children.length > 2) {
				syra_layoutUpdater.newLoadChildItems(section.layoutContent.loadChildItem(null, {
					$layoutType: "stack",
					$items: []
				}), children, 1);
			} else {
				if (children.length == 2) {
					section.layoutContent.loadChildItem(children[1]);
				}
			}
		}
		syra_dom.removeChild(prevLayout);
	}
};

function _selectColChoice(box, designedItem, colCount) {
	var css = "s-aw-cols-link " + (designedItem.$designLevel == "field" ? "" : "s-aw-row-box");
	var keys = Object.keys(box.colChoices);
	for (var ii = 0, jj = keys.length; ii < jj; ii++) {
		box.colChoices[keys[ii]].link.className = css + (colCount == keys[ii] ? " s-selected" : "");
	}
}

exports.load = function(designer) {
	designer.addLandingPageModels = function(slot, designedItem) {
		this.countTemplates = {
			2: ["2-1", "2-2"],
			3: ["3-1", "3-2", "3-3", "3-4"],
			4: ["4-1", "4-2", "4-3"],
			5: ["5-1", "5-2", "5-3"],
			6: ["6-1", "6-2"]
		};
		this.defaultTemplates = ["rows", "3cols", "4cols", "mixt"];
		var colCount = designedItem.layoutContent.getColumnsCount();
		syra_dom.empty(slot);
		var count = designedItem.layoutContent.getFields().length;
		if (count > 1) {
			var templates = this.countTemplates[count] || this.defaultTemplates;
			for (var ii = 0, jj = templates.length; ii < jj; ii++) {
				_addModelBtn(this, templates[ii], slot).$fieldsCount = count;
			}
		}
	};
	designer.addPageModelLinks = function(box, designedItem, colCount) {
		if (!box.modelSlot) {
			box.modelSlot = document.createElement("div");
			box.body.appendChild(box.modelSlot);
		}
		syra_dom.empty(box.modelSlot);
		if (!designedItem.isField) {
			if (designedItem.$designLevel == "article") {
				var slot = document.createElement("div");
				_addModelBtn(this, "smart", slot);
				_addModelBtn(this, "factory", slot);
				box.modelSlot.appendChild(slot);
			}
			var slot = document.createElement("div");
			if (designedItem.$designLevel == "article" || designedItem.$designLevel == "section") {
				_addModelBtn(this, "tabs", slot);
			}
			_addModelBtn(this, "rows", slot);
			_addModelBtn(this, "3cols", slot);
			_addModelBtn(this, "4cols", slot);
			_addModelBtn(this, "mixt", slot);
			box.modelSlot.appendChild(slot);
		}
		this.addColsCountButtons(box, designedItem, colCount, 6);
	};
	designer.addColsCountButtons = function(box, designedItem, colCount, max) {
		if (!box.colChoices) {
			box.colChoices = {};
			var slot = document.createElement("div");
			slot.className = "s-aw-cols";
			slot.textContent = syra_local.aw_columns;
			for (var ii = 1; ii <= max; ii++) {
				box.colChoices[ii] = syra_button.add({
					parent: designer,
					slot: slot,
					text: ii,
					css: "s-aw-template",
					btnclick: designer.onTemplateClick,
					$colCount: ii,
					$modelType: "colcount"
				});
			}
			box.body.appendChild(slot);
		}
		_selectColChoice(box, designedItem, colCount);
	};

	designer.onTemplateClick = function() {
		var designer = this.parent;
		if (designer.sectionCards && this.$modelType == "colcount") {
			designer.designedArticle.applyDesignMeta({
				$cardsByRowCount: this.$colCount
			}, true);
			_selectColChoice(designer.sectionCards, designer.designedArticle, this.$colCount);
			designer.endArticleUpdate();
		} else {
			var selectItem;
			var designedItem = designer.designedItem;
			var awLayout = designer.awLayout;
			if (this.$fieldsCount) {
				selectItem = designer.designedArticle.layoutContent.getOpenedTab();
				var funcName;
				var templates = designer.countTemplates[this.$fieldsCount];
				if (templates) {
					funcName = this.$fieldsCount;
				} else {
					templates = designer.defaultTemplates;
					funcName = this.$model;
				}
				var fields = selectItem.layoutContent.getFields();
				syra_layoutUpdater.extractItems(selectItem.layoutContent, fields);
				_builder[funcName](selectItem, this.$model, fields);
			}
			if (!selectItem) {
				switch (this.$modelType) {
					case "modelTab":
						if (designedItem.layoutParent.$layoutType != this.$layoutType) {
							if (designedItem.layoutParent.items.length == 1) {
								syra_layoutUpdater.convertToNewLayout(designedItem.layoutParent, {
									$layoutType: this.$layoutType
								});
							} else {
								var layoutRoot = designedItem.layoutParent;
								if (this.$layoutType != "stack") {
									var layoutTab = syra_layoutUpdater.wrapIntack(layoutRoot, [designedItem], layoutRoot.$layout.$items.indexOf(designedItem.$item));
									syra_layoutUpdater.convertToNewLayout(layoutTab, {
										$layoutType: "tabs"
									});
								}
							}
							selectItem = designedItem;
						}
						break;
					case "colcount":
						selectItem = _convertByColumns(this.$colCount, designedItem, awLayout);
						break;
					case "model_smart":
						_applyArticleSmartModel(designer);
						selectItem = designedItem;
						break;
					case "model_factory":
						if (designer.designedArticle.page == designer.designedArticle) {
							_applyPageFactoryModel(designer.designedArticle);
						} else {
							_applyCardListFactoryModel(designer.designedArticle);
						}
						selectItem = designedItem;
						break;
					case "model_3cols":
						var item = designer.getDesignedCardItem ? designer.getDesignedCardItem() : designedItem;
						_splitInColumns(item, null, 3);
						selectItem = designedItem;
						break;
					case "model_4cols":
						var item = designer.getDesignedCardItem ? designer.getDesignedCardItem() : designedItem;
						_splitInColumns(item, null, 4);
						selectItem = designedItem;
						break;
					case "model_mixt":
						var item = designer.getDesignedCardItem ? designer.getDesignedCardItem() : designedItem;
						_builder.mixt(item);
						selectItem = designedItem;
						break;
					default:
						if (this.$modelType.indexOf("model_") == 0) {
							var item = designer.getDesignedCardItem ? designer.getDesignedCardItem() : designedItem;
							_applySectionModel(item, this.$modelType);
							selectItem = designedItem;
						}
						break;
				}
			}
			selectItem && designer.endArticleUpdate(selectItem, true);
		}
	};
};
