// syracuse-ui/lib/authoring/tools/templatesTool.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function TemplatesTool() {}

exports.TemplatesTool = helpers.defineClass(TemplatesTool, null, {
	load: function(designer) {
		this.designer = designer;
		if (this.designer.awArticle.isDashBoard) {
			this.countTemplates = {
				2: ["2-1", "2-2"],
				3: ["3-1", "3-2", "3-3", "3-4"],
				4: ["4-1", "4-2", "4-3"],
				5: ["5-1", "5-2", "5-3"],
				6: ["6-1", "6-2"]
			};
			this.defaultTemplates = ["rows", "3cols", "4cols", "mixt"];
		}
	},
	dispose: function() {
		this.designer = null;
	},
	_applyArticleSmartModel: function(article) {
		var children = article.layoutContent.extractItems();
		article.layoutContent.clearContent();
		var prevLayout = article.layoutContent._item;
		article.$item.$layout = {
			$layoutType: "stack",
			$items: []
		};
		article.renderLayoutContent();
		article.layoutContent.clearContent();
		if (children.length > 0) {
			article.layoutContent.ensureSection(children[0]);
			if (children.length > 1) {
				if (children.length > 2) {
					var tabs = article.layoutContent.loadChildItem(null, {
						$layoutType: "tabs",
						$items: []
					});
					for (var ii = 1, jj = children.length - 1; ii < jj; ii++) {
						tabs.ensureSection(children[ii]);
					}
				}
				article.layoutContent.ensureSection(children[children.length - 1]);
			}
		}
		children = article.layoutContent.getItems();
		for (var ii = 0, jj = children.length; ii < jj; ii++) {
			this._applySectionSmartModel(children[ii]);
		}
		document.site.removeDomChild(prevLayout);
	},
	_applySectionSmartModel: function(section) {
		var children = section.layoutContent.extractItems();
		section.layoutContent.clearContent();
		var prevLayout = section.layoutContent._item;
		if (children.length > 0) {
			var $rows = [];
			var colIndex = 0;
			for (var ii = 0, jj = children.length; ii < jj; ii++) {
				colIndex++;
				if (children[ii].isSection) {
					var subChildren = children[ii].layoutContent.getItems();
					if (subChildren && subChildren.length > 0) {
						if (subChildren[0].$field && subChildren[0].$field.$type == "application/x-array") {
							$rows.push({
								$layoutType: colIndex > 1 ? "row" : "stack",
								$items: []
							});
							colIndex = 1;
						}
					}
				} else {
					if (children[ii].$field && children[ii].$field.$type == "application/x-array") {
						if (colIndex > 1) {
							$rows.push({
								$layoutType: "row",
								$items: []
							});
						}
						$rows.push({
							$layoutType: "stack",
							$items: []
						});
						colIndex = 0;
					}
				}
				if (colIndex == 3) {
					$rows.push({
						$layoutType: "row",
						$items: []
					});
					colIndex = 0;
				}
			}
			if (colIndex) {
				$rows.push({
					$layoutType: colIndex > 1 ? "row" : "stack",
					$items: []
				});
			}
			if ($rows.length == 1) {
				section.$item.$layout = $rows[0];
				section.renderLayoutContent();
				section.layoutContent.removeSpaceBox(true);
				section.layoutContent.newLoadChildItems(children);
			} else {
				section.$item.$layout = {
					$layoutType: "stack",
					$items: $rows
				};
				section.renderLayoutContent();
				var rowIndex = colIndex = 0;
				for (var ii = 0, jj = children.length; ii < jj; ii++) {
					colIndex++;
					if (children[ii].isSection) {
						var subChildren = children[ii].layoutContent.getItems();
						if (subChildren && subChildren.length > 0) {
							if (subChildren[0].$field && subChildren[0].$field.$type == "application/x-array") {
								rowIndex++;
								colIndex = 1;
							}
						}
					} else {
						if (children[ii].$field && children[ii].$field.$type == "application/x-array") {
							if (colIndex > 1) {
								rowIndex++;
							}
							colIndex = 3;
						}
					}
					section.layoutContent.items[rowIndex].removeSpaceBox(true);
					section.layoutContent.items[rowIndex].loadChildItem(children[ii]);
					if (colIndex == 3) {
						rowIndex++;
						colIndex = 0;
					}
				}
			}
		}

		children = section.layoutContent.getItems();
		for (var ii = 0, jj = children.length; ii < jj; ii++) {
			if (children[ii].isSection) {
				this._applySectionSmartModel(children[ii]);
			}
		}
		document.site.removeDomChild(prevLayout);
	},
	_applyPageFactoryModel: function(page) {
		var children = page.layoutContent.extractItems();
		page.layoutContent.clearContent();
		var prevLayout = page.layoutContent._item;
		page.$item = helpers.object.clone(page.ensureDefaultArticle(page.$prototype.$article, page.$prototype), true);
		page.reloadLayout(page.$item);
	},
	_createLayout: function(section, $newLayout) {
		section.layoutContent.clearContent();
		var prevLayout = section.layoutContent._item;
		section.$item.$layout = $newLayout;
		section.renderLayoutContent();
		section.layoutContent.removeSpaceBox(true);
		return prevLayout;
	},
	splitInColumns: function(section, children, maxCol) {
		children = children || section.layoutContent.extractItems();
		var prevLayout = this._createLayout(section, {
			$layoutType: "row",
			$widths: section.page.layoutValidator.getDefaultWidths(maxCol),
			$items: []
		});
		var row = 0,
			colIndex = 0;
		for (var ii = 0, jj = children.length; ii < jj; ii++) {
			if (!children[ii].disposed) {
				var layout;
				switch (row) {
					case 0:
						layout = section.layoutContent;
						break;
					case 1:
						layout = section.layoutContent.wrapIntack([section.layoutContent.items[colIndex]], colIndex);
						break;
					default:
						layout = section.layoutContent.items[colIndex];
						break;
				}
				colIndex++;
				layout.loadChildItem(children[ii]);
				if (colIndex == maxCol) {
					colIndex = 0;
					row++;
				}
			}
		}
		document.site.removeDomChild(prevLayout);
	},
	_applySectionColumnsModel: function(section, children, maxCols) {
		section.$item.$layout = {
			$layoutType: "row",
			$items: []
		};
		var width = Math.ceil(100 / maxCols);
		var widths = [];
		for (var ii = 0; ii < maxCols; ii++) {
			widths.push(width);
		}
		section.$item.$layout.$widths = widths.join(",");
		if (children.length > maxCols) {
			for (var ii = 0; ii < maxCols; ii++) {
				section.$item.$layout.$items.push({
					$layoutType: "stack",
					$items: []
				});
			}
		}
		section.renderLayoutContent();
		if (children.length > maxCols) {
			var colIndex = 0;
			for (var ii = 0, jj = children.length; ii < jj; ii++) {
				section.layoutContent.items[colIndex++].loadChildItem(children[ii]);
				if (colIndex == maxCols) {
					colIndex = 0;
				}
			}
		} else {
			section.layoutContent.clearContent();
			section.layoutContent.newLoadChildItems(children);
		}
	},
	_applySectionModel: function(section, $model) {
		var children = section.layoutContent.extractItems();
		var prevLayout = section.layoutContent._item;
		section.layoutContent.clearContent();
		switch ($model) {
			case "$model_rows":
				section.$item.$layout = {
					$layoutType: "stack",
					$items: []
				};
				section.renderLayoutContent();
				section.layoutContent.newLoadChildItems(children);
				break;
			case "$model_3cols":
				this._applySectionColumnsModel(section, children, 3);
				break;
			case "$model_4cols":
				this._applySectionColumnsModel(section, children, 4);
				break;
			case "$model_tabs":
				section.$item.$layout = {
					$layoutType: "tabs",
					$items: []
				};
				section.renderLayoutContent();
				section.layoutContent.clearContent();
				for (var ii = 0, jj = children.length; ii < jj; ii++) {
					section.layoutContent.ensureSection(children[ii]);
				}
				break;
		}
		document.site.removeDomChild(prevLayout);
	},
	_addModelLink: function($model, slot, isTextIcon) {
		var link = document.createElement("a");
		link.title = this.designer.localize["aw_model_" + $model];
		link.setAttribute("data-s-picker", link.syraPickerType = "$model_" + $model);
		link.syraTemplate = $model;
		link.className = "s-aw-template";
		var img = document.createElement("div");
		img.className = "s-aw-template-icon";
		img.style.backgroundImage = "url('/syracuse-ui/themes/desktop/sage/images/authoring/s-template-" + $model + ".png')";
		link.appendChild(img);
		if (isTextIcon) {
			var label = document.createElement("div");
			label.className = "s-aw-template-label";
			label.textContent = link.title;
			link.appendChild(label);
		} else {
			link.className += " s-icon";
		}
		return slot.appendChild(link);
	},
	addModelLinks: function(section, awItem, colCount, awLayout) {
		if (this.designer.awArticle.isDashBoard) {
			var colCount = awItem.layoutContent.getColumnsCount();
			var fields = awItem.layoutContent.getFields();
			section.dataStore = section.dataStore || {};
			if (!section.dataStore.modelSlot) {
				section.dataStore.modelSlot = document.createElement("div");
				section.body.appendChild(section.dataStore.modelSlot);
			}
			document.site.emptyDom(section.dataStore.modelSlot);
			var count = fields.length;
			if (count > 1) {
				//for (var mm = 2; mm <= 6; mm++) {
				//count = mm;
				var templates = this.countTemplates[count] || this.defaultTemplates;
				var slot = document.createElement("div");
				slot.className = "s-aw-template-row";
				for (var ii = 0, jj = templates.length; ii < jj; ii++) {
					this._addModelLink(templates[ii], slot).syraFieldsCount = count;
				}
				section.dataStore.modelSlot.appendChild(slot);
				//}
				this.addColsCountLinks(section, awItem, colCount, 3);
			}
		} else {
			section.dataStore = section.dataStore || {};
			if (!section.dataStore.modelSlot) {
				section.dataStore.modelSlot = document.createElement("div");
				section.body.appendChild(section.dataStore.modelSlot);
			}
			document.site.emptyDom(section.dataStore.modelSlot);
			if (!awItem.isField) {
				if (awItem.$designLevel == "article") {
					var slot = document.createElement("div");
					slot.className = "s-aw-template-row";
					this._addModelLink("smart", slot, true);
					this._addModelLink("factory", slot, true);
					section.dataStore.modelSlot.appendChild(slot);
				}
				var slot = document.createElement("div");
				slot.className = "s-aw-template-row";
				if (awItem.$designLevel == "article" || awItem.$designLevel == "section") {
					this._addModelLink("tabs", slot);
				}
				this._addModelLink("rows", slot);
				this._addModelLink("3cols", slot);
				this._addModelLink("4cols", slot);
				this._addModelLink("mixt", slot);
				section.dataStore.modelSlot.appendChild(slot);
			}
			this.addColsCountLinks(section, awItem, colCount, 6);
		}
	},
	addColsCountLinks: function(section, awItem, colCount, max) {
		section.dataStore = section.dataStore || {};
		if (!section.dataStore.colChoices) {
			section.dataStore.colChoices = {};
			var row = document.createElement("div");
			row.className = "s-aw-cols";
			var cell = document.createElement("div");
			cell.className = "s-aw-cols-label";
			cell.textContent = this.designer.localize.aw_columns;
			row.appendChild(cell);

			for (var ii = 1; ii <= max; ii++) {
				var link = document.createElement("a");
				link.textContent = link.$syraColCount = ii;
				link.setAttribute("data-s-picker", link.syraPickerType = "$colcount");
				row.appendChild(section.dataStore.colChoices[ii] = link);
			}
			section.body.appendChild(row);
		}
		this.selectColChoice(section, awItem, colCount);
	},
	selectColChoice: function(section, awItem, colCount) {
		var css = "s-aw-cols-link " + (awItem.$designLevel == "field" ? "" : "s-aw-row-box");
		var keys = Object.keys(section.dataStore.colChoices);
		for (var ii = 0, jj = keys.length; ii < jj; ii++) {
			section.dataStore.colChoices[keys[ii]].className = css + (colCount == keys[ii] ? " s-selected" : "");
		}
	},
	fillTabChoice: function(section, awItem) {
		var title = this.designer.localize.aw_showAs;
		var value = awItem.isTabLayout ? "tabs" : "stack";
		var isLinkDisabled = false;
		var links = ["tabs", "stack"];
		section.dataStore = section.dataStore || {};

		if (!section.dataStore.tabChoice) {
			section.dataStore.tabChoice = {};
			var row = document.createElement("div");
			row.className = "s-aw-tab-choices";
			var cell = document.createElement("div");
			cell.className = "s-aw-tab-choices-label";
			section.dataStore.tabChoice.label = cell;
			row.appendChild(cell);

			for (var ii = 0, jj = links.length; ii < jj; ii++) {
				var link = document.createElement("a");
				section.dataStore.tabChoice[link.$syraValue = links[ii]] = link;
				link.title = this.designer.localize["aw_section_" + link.$syraValue];
				link.className = "s-aw-tab-choice";
				link.setAttribute("data-s-picker", link.syraPickerType = "$boxTabChoice");
				var img = document.createElement("div");
				img.className = "s-aw-tab-choice-icon";
				img.style.backgroundImage = "url('/syracuse-ui/themes/desktop/sage/images/authoring/s-layout-" + link.$syraValue + ".png')";
				link.appendChild(img);
				row.appendChild(link);
			}
			section.body.appendChild(row);
		}

		section.dataStore.tabChoice.label.textContent = title;
		for (var ii = 0, jj = links.length; ii < jj; ii++) {
			var link = section.dataStore.tabChoice[links[ii]];
			link.className = "s-aw-tab-choice" + (value == links[ii] ? " s-selected" : "");
			document.site.disableItem(link, isLinkDisabled);
		}
	},
	onTemplateChoiceClick: function(picker, awItem, awLayout) {
		var selectItem = this.designer.onTemplateChoiceClick ? this.designer.onTemplateChoiceClick(picker, awItem, awLayout) : null;
		if (!selectItem) {
			switch (picker.syraPickerType) {
				case "$boxTabChoice":
					if (awItem.layoutParent.$layoutType != picker.$syraValue) {
						if (awItem.layoutParent.items.length == 1) {
							awItem.layoutParent.convertToNewLayout({
								$layoutType: picker.$syraValue
							});
						} else {
							var layoutRoot = awItem.layoutParent;
							if (picker.$syraValue != "stack") {
								var layoutTab = layoutRoot.wrapIntack([awItem], layoutRoot.$layout.$items.indexOf(awItem.$item));
								layoutTab.convertToNewLayout({
									$layoutType: "tabs"
								});
							}
						}
						return awItem;
					}
					break;
				case "$colcount":
					return this.convertByColumns(picker, awItem, awLayout);
				case "$model_smart":
					this._applyArticleSmartModel(awItem.page);
					return awItem;
				case "$model_factory":
					this._applyPageFactoryModel(awItem.page);
					return awItem;
				case "$model_3cols":
					this.splitInColumns(awItem, null, 3);
					return awItem;
				case "$model_4cols":
					this.splitInColumns(awItem, null, 4);
					return awItem;
				case "$model_mixt":
					this.build_mixt(awItem);
					return awItem;
				default:
					if (picker.syraPickerType.indexOf("$model_") == 0) {
						this._applySectionModel(awItem, picker.syraPickerType);
						return awItem;
					}
					break;
			}
		}
		return selectItem;
	},
	convertByColumns: function(picker, awItem, awLayout) {
		var layout = awLayout ? awLayout : awItem.layoutContent;
		if (awItem.$designLevel == "field") {
			layout = awItem.layoutParent.wrapIntack([awItem], awItem.layoutParent.items.indexOf(awItem));
			layout.removeSpaceBox(true);
		}
		var $widths = ["100", "50,50", "33,33,33", "25,25,25,25", "20,20,20,20,20", "17,17,16,16,17,17"][picker.$syraColCount - 1];
		layout.convertToNewLayout({
			$layoutType: $widths == "100" ? "stack" : "row",
			$widths: $widths
		});
		return awLayout ? awLayout : awItem.layoutContent;
	},
	build_2: function(awItem, template, fields) {
		var prevLayout = this._createLayout(awItem, (template == "2-1") ? {
			$layoutType: "stack",
			$items: []
		} : {
			$layoutType: "row",
			$widths: "50,50",
			$items: []
		});
		awItem.layoutContent.newLoadChildItems(fields);
		document.site.removeDomChild(prevLayout);
	},
	build_3: function(awItem, template, fields) {
		var prevLayout;
		switch (template) {
			case "3-1":
				prevLayout = this._createLayout(awItem, {
					$layoutType: "stack",
					$items: [{
						$layoutType: "row",
						$widths: "50,50",
						$items: []
					}]
				});
				awItem.layoutContent.items[0].removeSpaceBox(true);
				awItem.layoutContent.items[0].newLoadChildItems(fields.splice(0, 2));
				awItem.layoutContent.newLoadChildItems(fields);
				break;
			case "3-2":
				prevLayout = this._createLayout(awItem, {
					$layoutType: "row",
					$widths: "50,50",
					$items: []
				});
				awItem.layoutContent.loadChildItem(null, {
					$layoutType: "stack",
					$items: []
				}).newLoadChildItems(fields.splice(0, 2));
				awItem.layoutContent.newLoadChildItems(fields);
				break;
			case "3-3":
				prevLayout = this._createLayout(awItem, {
					$layoutType: "row",
					$widths: "50,50",
					$items: []
				});
				awItem.layoutContent.newLoadChildItems(fields.splice(0, 1));
				awItem.layoutContent.loadChildItem(null, {
					$layoutType: "stack",
					$items: []
				}).newLoadChildItems(fields);
				break;
			case "3-4":
				prevLayout = this._createLayout(awItem, {
					$layoutType: "stack",
					$items: []
				});
				awItem.layoutContent.newLoadChildItems(fields.splice(0, 1));
				var row = awItem.layoutContent.loadChildItem(null, {
					$layoutType: "row",
					$widths: "50,50",
					$items: []
				});
				row.removeSpaceBox(true);
				row.newLoadChildItems(fields);
				break;
		}
		document.site.removeDomChild(prevLayout);
	},
	build_4: function(awItem, template, fields) {
		var prevLayout;
		switch (template) {
			case "4-1":
				prevLayout = this._createLayout(awItem, {
					$layoutType: "stack",
					$items: [{
						$layoutType: "row",
						$widths: "50,50",
						$items: []
					}, {
						$layoutType: "row",
						$widths: "50,50",
						$items: []
					}]
				});
				var row = awItem.layoutContent.items[0];
				row.removeSpaceBox(true);
				row.newLoadChildItems(fields.splice(0, 2));
				row = awItem.layoutContent.items[1];
				row.removeSpaceBox(true);
				row.newLoadChildItems(fields);
				break;
			case "4-2":
				prevLayout = this._createLayout(awItem, {
					$layoutType: "stack",
					$items: []
				});
				awItem.layoutContent.newLoadChildItems(fields.splice(0, 1));
				var row = awItem.layoutContent.loadChildItem(null, {
					$layoutType: "row",
					$widths: "33,33,33",
					$items: []
				});
				row.removeSpaceBox(true);
				row.newLoadChildItems(fields);
				break;
			case "4-3":
				prevLayout = this._createLayout(awItem, {
					$layoutType: "stack",
					$items: [{
						$layoutType: "row",
						$widths: "33,33,33",
						$items: []
					}]
				});
				var row = awItem.layoutContent.items[0];
				row.removeSpaceBox(true);
				row.newLoadChildItems(fields.splice(0, 3));
				awItem.layoutContent.newLoadChildItems(fields);
				break;

		}
		document.site.removeDomChild(prevLayout);
	},
	build_5: function(awItem, template, fields) {
		var prevLayout;
		switch (template) {
			case "5-1":
				prevLayout = this._createLayout(awItem, {
					$layoutType: "row",
					$widths: "50,50",
					$items: [{
						$layoutType: "stack",
						$items: []
					}, {
						$layoutType: "stack",
						$items: []
					}]
				});
				awItem.layoutContent.items[0].newLoadChildItems(fields.splice(0, 2));
				awItem.layoutContent.items[1].newLoadChildItems(fields);
				break;
			case "5-2":
				prevLayout = this._createLayout(awItem, {
					$layoutType: "stack",
					$items: [{
						$layoutType: "row",
						$widths: "33,33,33",
						$items: []
					}, {
						$layoutType: "row",
						$widths: "50,50",
						$items: []
					}]
				});
				var row = awItem.layoutContent.items[0];
				row.removeSpaceBox(true);
				row.newLoadChildItems(fields.splice(0, 3));
				row = awItem.layoutContent.items[1];
				row.removeSpaceBox(true);
				row.newLoadChildItems(fields);
				break;
			case "5-3":
				prevLayout = this._createLayout(awItem, {
					$layoutType: "stack",
					$items: []
				});
				awItem.layoutContent.newLoadChildItems(fields.splice(0, 1));
				var row = awItem.layoutContent.loadChildItem(null, {
					$layoutType: "row",
					$widths: "50,50",
					$items: []
				});
				row.removeSpaceBox(true);
				row.newLoadChildItems(fields.splice(0, 2));
				var row = awItem.layoutContent.loadChildItem(null, {
					$layoutType: "row",
					$widths: "50,50",
					$items: []
				});
				row.removeSpaceBox(true);
				row.newLoadChildItems(fields);
				break;

		}
		document.site.removeDomChild(prevLayout);
	},
	build_6: function(awItem, template, fields) {
		var prevLayout;
		switch (template) {
			case "6-1":
				prevLayout = this._createLayout(awItem, {
					$layoutType: "stack",
					$items: [{
						$layoutType: "row",
						$widths: "50,50",
						$items: []
					}, {
						$layoutType: "row",
						$widths: "50,50",
						$items: []
					}, {
						$layoutType: "row",
						$widths: "50,50",
						$items: []
					}]
				});
				var row = awItem.layoutContent.items[0];
				row.removeSpaceBox(true);
				row.newLoadChildItems(fields.splice(0, 2));
				row = awItem.layoutContent.items[1];
				row.removeSpaceBox(true);
				row.newLoadChildItems(fields.splice(0, 2));
				row = awItem.layoutContent.items[2];
				row.removeSpaceBox(true);
				row.newLoadChildItems(fields.splice(0, 2));
				break;
			case "6-2":
				prevLayout = this._createLayout(awItem, {
					$layoutType: "stack",
					$items: [{
						$layoutType: "row",
						$widths: "33,33,33",
						$items: []
					}, {
						$layoutType: "row",
						$widths: "33,33,33",
						$items: []
					}]
				});
				var row = awItem.layoutContent.items[0];
				row.removeSpaceBox(true);
				row.newLoadChildItems(fields.splice(0, 3));
				row = awItem.layoutContent.items[1];
				row.removeSpaceBox(true);
				row.newLoadChildItems(fields.splice(0, 3));
				break;

		}
		document.site.removeDomChild(prevLayout);
	},
	build_rows: function(awItem, template, fields) {
		document.site.removeDomChild(prevLayout);
	},
	build_3cols: function(awItem, template, fields) {
		this.splitInColumns(awItem, fields, 3);
	},
	build_4cols: function(awItem, template, fields) {
		this.splitInColumns(awItem, fields, 4);
	},
	build_mixt: function(section, template, children) {
		children = children || section.layoutContent.extractItems();
		var prevLayout = section.layoutContent._item;
		section.layoutContent.clearContent();
		section.$item.$layout = {
			$layoutType: "row",
			$widths: "50,50",
			$items: []
		};
		section.renderLayoutContent();
		section.layoutContent.clearContent();
		if (children.length) {
			section.layoutContent.loadChildItem(children[0]);
			if (children.length > 2) {
				section.layoutContent.loadChildItem(null, {
					$layoutType: "stack",
					$items: []
				}).newLoadChildItems(children, 1);
			} else {
				section.layoutContent.loadChildItem(children[1]);
			}
		}
		document.site.removeDomChild(prevLayout);
	}
});
