// syracuse-ui/lib/authoring/tools/uiSwitchTool.js
"use strict";

function _onDeleteLayoutClick() {
	if (!syra_dd.ddAgent) {
		syra_dd.cancel();
		var designer = this.parent;
		var spaceBox = this.spaceBox;
		setTimeout(function() {
			var layout = spaceBox.layoutParent;
			var layoutParent = layout.layoutParent;
			var deletedIndex;
			if (spaceBox) {
				deletedIndex = layout.items.indexOf(spaceBox);
				layout.removeItem(spaceBox, true);
			}
			if (layout.items.length > 0 && layout.isRow) {
				syra_layoutUpdater.deleteRowColumn(layout, deletedIndex);
				designer.endArticleUpdate(layout, true);
			} else {
				if (layoutParent) {
					layoutParent.removeItem(layout, true);
					layoutParent.isRow && syra_layoutUpdater.deleteRowColumn(layoutParent, deletedIndex);
				} else {
					layoutParent = layout.boxParent.layoutParent;
					layoutParent.removeItem(layout.boxParent, true, layoutParent.isRow);
				}
				designer.endArticleUpdate(layoutParent, true);
			}
		}, 10);
	}
}

exports.load = function(designer) {
	designer.toggleUIDesign = function(item, enable, disposingDesigner) {
		var site = syra_site;
		var isLightMode;
		if (this.$viewType) {
			isLightMode = this.$viewType == "viewTypeStructure";
			if (this.$viewType == "viewTypePreview") {
				enable = false;
			}
		}
		if (this.designedArticle.isRecordArticle) {
			if (item.articleParent && !((item.articleParent == this.designedArticle.list) ||
				(item.articleParent == this.designedArticle))) {
				return;
			}
		} else {
			if (item.articleParent && item.articleParent != this.designedArticle.page) {
				return;
			}
		}
		if (item.$isAuthoringEnabled === false) {
			return;
		}
		item.isDesigning = enable;
		if (item.isSpaceBox) {
			if (enable) {
				if (!item.deleteBtn) {
					if (item.boxParent.$designLevel == "block") {
						item.$designLevel = "field";
					} else {
						var children = item.boxParent.layoutContent ? item.boxParent.layoutContent.getItems() : [];
						if (children.length > 0) {
							item.$designLevel = children[0].$designLevel;
						} else {
							item.$designLevel = item.boxParent.$designLevel == "section" ? "block" : "section";
						}
					}
					item.deleteBtn = syra_menus.button.add({
						parent: this,
						slot: item.body,
						text: syra_local.aw_deleteColumn,
						iconOnly: true,
						css: "s-aw-del-layout",
						fontIcon: "delete",
						spaceBox: item,
						btnclick: _onDeleteLayoutClick
					});

				}
			} else {
				if (item.deleteBtn) {
					syra_menus.button.remove(item.deleteBtn);
					item.deleteBtn = null;
				}
				syra_dom.toggleClass(item.layoutSlot, "s-aw-designed", false);
				syra_dom.toggleClass(item.layoutSlot, "s-aw-over", false);
			}
			if (!site.isTabletDevice) {
				item.layoutSlot.syraDragSpot = enable ? item.id : null;
			}
			syra_dom.toggleClass(item.layoutSlot, "s-aw-item", enable);
		}
		if (item.isTabLayout) {
			if (!item.isLayout && item.isSection && !item.loaded && item.$item.$opened !== false) {
				item.renderLayoutContent();
				item.loaded = true;
			}
		}
		var domItem = item.dataSlot || item.domItem;
		if (domItem) {
			//no domitem for layout
			if (item.isSection) {
				if (isLightMode) {
					if (!item.childrenSection || item.childrenSection.length == 0) {
						if (item.layoutContent && item.layoutContent.domItem) {
							item.$isLayoutContentLightHidden = true;
							item.layoutContent.domItem.style.display = "none";
						}
					}
				} else {
					if (item.$isLayoutContentLightHidden) {
						item.layoutContent.domItem.style.display = "";
					}
				}
			}
			if (item.isTabSection) {
				if (!enable) {
					syra_dom.toggleClass(item.header, "s-aw-designed", false);
					syra_dom.toggleClass(item.header, "s-aw-over", false);
					syra_dom.toggleClass(item.body, "s-aw-designed", false);
					syra_dom.toggleClass(item.body, "s-aw-over", false);
				}
				syra_dom.toggleClass(item.header, "s-aw-item", enable);
				syra_dom.toggleClass(item.body, "s-aw-item", enable);
				if (!site.isTabletDevice) {
					item.header.syraDragSpot = enable ? item.id : null;
					item.body.syraDragSpot = enable ? item.id : null;
				}
				syra_dom.toggleClass(item.header, "s-aw-light", enable && isLightMode);
				syra_dom.toggleClass(item.body, "s-aw-light", enable && isLightMode);
			}
			if (domItem && !enable) {
				if (item.body) {
					syra_dom.toggleClass(item.body, "s-aw-over", false);
					syra_dom.toggleClass(item.body, "s-aw-designed", false);
				}
				if (item.header) {
					syra_dom.toggleClass(item.header, "s-aw-over", false);
					syra_dom.toggleClass(item.header, "s-aw-designed", false);
				}
				syra_dom.toggleClass(item.layoutSlot, "s-aw-designed", false);
				syra_dom.toggleClass(item.layoutSlot, "s-aw-over", false);
				syra_dom.toggleClass(domItem, "s-aw-designed", false);
				syra_dom.toggleClass(domItem, "s-aw-over", false);
			}
			if (item.isLayout) {
				if (item.$layout.$layoutType == "row") {
					item.domItem.syraItem = item.id;
					var children = item.domItem.children;
					item.page.isLandingPage && syra_dom.toggleClass(item.domItem, "s-aw-ldp-row", enable);
					syra_dom.toggleClass(item.domItem, "s-aw-item", enable);
					if (!site.isTabletDevice) {
						item.domItem.syraDragSpot = enable ? item.id : null;
					}
					if (item.layoutSlot) {
						syra_dom.toggleClass(item.layoutSlot, "s-aw-item", enable);
						if (!site.isTabletDevice) {
							item.layoutSlot.syraDragSpot = enable ? item.id : null;
						}
					}
				} else {
					if (item.layoutParent && item.layoutParent.isRow) {
						item.page.isLandingPage && syra_dom.toggleClass(item.layoutSlot, "s-aw-ldp-row-col", enable);
						syra_dom.toggleClass(item.layoutSlot, "s-aw-item", enable);
						if (!site.isTabletDevice) {
							item.layoutSlot.syraDragSpot = enable ? item.id : null;
						}
					}
				}
				if (!enable) {
					item.domItem.style.display = "";
					delete item.domItem.syraItem;
				}
			} else {
				if (item.layoutSlot) {
					if (item.page.isLandingPage) {
						if (item.layoutParent && item.layoutParent.isRow) {
							syra_dom.toggleClass(item.layoutSlot, "s-aw-ldp-row-col", enable);
						}
					}
					syra_dom.toggleClass(item.layoutSlot, "s-aw-item", enable);
					item.layoutSlot.syraItem = item.id;
					if (!site.isTabletDevice) {
						item.layoutSlot.syraDragSpot = enable ? item.id : null;
					}
				}
				syra_dom.toggleClass(domItem, "s-aw-item", enable);
				if (!site.isTabletDevice) {
					domItem.syraDragSpot = enable ? item.id : null;
				}
				syra_dom.toggleClass(domItem, "s-aw-light", enable && isLightMode);
				if (item.page != item) {
					var show = enable || !(item.$isHidden || item.isItemHidden || item.isInvisible);
					item.showItem && item.showItem(show);
				}

			}
		}

		item.toggleUIDesign && item.toggleUIDesign(enable, disposingDesigner);
		if (item.isLayout) {
			if (item.items) {
				for (var ii = 0, jj = item.items.length; ii < jj; ii++) {
					this.toggleUIDesign(item.items[ii], enable, disposingDesigner);
				}
			}
		} else {
			item.layoutContent && this.toggleUIDesign(item.layoutContent, enable, disposingDesigner);
		}
	};
	designer.toggleOverItem = function(item, isOver) {
		this.toggleItemCss(item, "s-aw-over", isOver);
	};
	designer.toggleItemCss = function(item, css, show) {
		if (item.domItem) {
			if (!item.isTabSection) {
				if (item.dataSlot) {
					syra_dom.toggleClass(item.dataSlot, css, show); //page
				} else {
					item.layoutSlot && syra_dom.toggleClass(item.layoutSlot, css, show);
				}
				item.isLayout && syra_dom.toggleClass(item.domItem, css, show);
			}
			item.domItem && syra_dom.toggleClass(item.domItem, css, show);
			item.header && syra_dom.toggleClass(item.header, css, show);
			item.body && syra_dom.toggleClass(item.body, css, show);
		}
	};
};
