// @sage/syracuse-ui/lib/scroll/list.js
"use strict";
var Scrollbar = require('@sage/syracuse-ui/lib/scroll/scrollbar').Scrollbar;

function List(config) {
	this.config = config || {};
	this.css = (this.config.css || "s-list-default");
	this.cssLine = this.config.cssLine || (this.css + "-li ");
	this._css_ul = "s-list-ul " + this.css + "-ul";
	this._css_li = "s-list-li " + this.cssLine;
	this._css_btn = this.css + "-btn-default";

	this.ul = config.ul || syra_dom.ul(this._css_ul);
	this.scrollview = this.ul;
	if (this.config.columns) {
		var width = Math.floor(100 / this.config.columns);
		this.scrollview = this.row = syra_dom.div(this.css + "-ul-row s-list-row");
		this.ul.className += " s-list-col";
		this.ul.style.width = width + "%";
		this.columns = [this.row.appendChild(this.ul)];
		for (var ii = 1; ii < this.config.columns; ii++) {
			var col = syra_dom.ul(this._css_ul + " s-list-col", this.row);
			col.style.width = width + "%";
			this.columns.push(col);
		}
		this.columnsIndex = 0;
	}
	if (this.config.scroll) {
		this.scrollSlot = syra_dom.div(this.css + "-scroll-slot");
		this.scrollBar = new Scrollbar({
			container: this.scrollSlot,
			scrollview: this.scrollview,
			scroll: this.config.scroll
		});
		this.scrollSlot.appendChild(this.scrollview);
	} else {
		this.scrollSlot = this.scrollview;
	}
	if (config.sortable) {
		config.sortable.root = this.scrollview;
		config.sortable.scrollview = this.scrollview;
		this.sortable = syra_sortable.add(config.sortable);
	}
	this._addNoChild();
	this.config.parentNode && config.parentNode.appendChild(this.scrollSlot);

}

List.prototype.empty = function() {
	var cols = this.columns || [this.ul];
	for (var ii = 0, jj = cols.length; ii < jj; ii++) {
		syra_dom.empty(cols[ii]);
	}
	//reset autocolumns
	if (this.config.autoColumns) {
		this.ul.className = this.ul.className.replace("s-list-col", "");
		delete this.row;
		this.scrollSlot.appendChild(this.scrollview = this.ul);
	}
	this._addNoChild();
};
List.prototype._addNoChild = function() {
	if (this.config.emptyText) {
		this.noChild = syra_dom.li(this._css_li, this.ul);
		syra_dom.span(this.css + "-empty", this.noChild).textContent = this.config.emptyText;
	}
};
List.prototype.remove = function(line) {
	syra_dom.remove(line);
	if (!this.ul.childNodes.length) {
		this._addNoChild();
	}
};
List.prototype.addLine = function(css, line) {
	var slot = this.ul;
	if (this.noChild) {
		syra_dom.remove(this.noChild);
		delete this.noChild;
	}
	if (this.columns) {
		slot = this.columns[this.columnsIndex];
		if (this.config.columns <= (++this.columnsIndex)) {
			this.columnsIndex = 0;
		}
	}
	if (!line) {
		line = syra_dom.li(this._css_li + (css || ""), slot);
	} else {
		slot.appendChild(line);
	}
	return line;
};
List.prototype.addAnchor = function(css, cssLine) {
	return syra_dom.a(css || this._css_btn, this.addLine(cssLine));
};
List.prototype.addReorder = function(line, handle) {
	var reorder = line.insertBefore(syra_dom.div(this.css + "-reorder"), line.firstChild);
	reorder.syraSortableHandle = handle;
	return reorder;
};
List.prototype.addButton = function(btn, cssLine) {
	btn.parent = btn.parent || syra_site;
	btn.slot = this.addLine(cssLine);
	btn.css = btn.css || this._css_btn;
	return syra_button.add(btn);
};
List.prototype.addItem = function(page, $item, cssLine) {
	return page.addItem(this.addLine(cssLine), $item);
};

List.prototype.scrollTo = function(node) {
	syra_site.scrollToItem(node, this.scrollview, true);
};
List.prototype.hide = function(hide) {
	syra_dom.hide(this.scrollSlot, hide);
};
List.prototype._applyAutoColumns = function(height) {
	if (!this.row && this.ul.clientHeight != this.ul.scrollHeight) {
		var parentNode = this.ul.parentNode;
		this.row = this.scrollview = syra_dom.div(this.css + "-ul-row s-list-row");
		this.ul.className += " s-list-col";
		this.ul.style.maxHeight = "";
		this.columns = [this.row.appendChild(this.ul)];
		var width = Math.floor(100 / this.config.autoColumns);
		var move = [
			[]
		];
		this.ul.style.width = width + "%";
		for (var ii = 1; ii < this.config.autoColumns; ii++) {
			var ul = syra_dom.ul(this._css_ul + " s-list-col", this.row);
			ul.style.width = width + "%";
			this.columns.push(ul);
			move.push([]);
		}
		var nodes = this.ul.childNodes;
		var columnsIndex = 0;
		for (var ii = 0, jj = nodes.length; ii < jj; ii++) {
			move[columnsIndex].push(nodes[ii]);
			if (this.config.autoColumns <= (++columnsIndex)) {
				columnsIndex = 0;
			}
		}
		for (var ii = 0, jj = move.length; ii < jj; ii++) {
			while (move[ii].length) {
				this.columns[ii].appendChild(move[ii].shift());
			}
		}
		this.scrollBar.setScrollView(this.row);
		parentNode.appendChild(this.row);
	}
};
List.prototype.resize = function(height) {
	this.scrollview.style.maxHeight = height + "px";
	this.config.autoColumns && this._applyAutoColumns();
	this.scrollBar && this.scrollBar.resize();
};

List.prototype.dispose = function() {
	this.sortable && this.sortable.dispose();
	this.scrollBar && this.scrollBar.dispose();
	syra_site.disposeObject(this);
};

exports.List = List;
