// syracuse-ui/lib/field/schedule/jqueryGantt.js
"use strict";

var helpers = require('syracuse-core').helpers;
var datetime = require('syracuse-core').types.datetime;
var date = require('syracuse-core').types.date;
var gantt = require('syracuse-ui/deps/jQuery.Gantt/js/jquery.fn.gantt');
require('syracuse-ui/deps/jQuery.popover/jquery.popover-1.1.0');

var forEachKey = helpers.object.forEachKey;
var objectCopy = helpers.object.copy;
var objectClone = helpers.object.clone;
var _localize = {};
// var _dateFormat = 'yyyy-mm-dd';
exports.setDataValue = function(value) {
	this.currentValue = value;

	if (!this._initialized) {
		var gantt = this.$$widgetBox.gantt({
			source: [{
				name: "Sprint 0",
				desc: "Analysis",
				values: [{
					from: "/Date(1320192000000)/",
					to: "/Date(1322401600000)/",
					label: "Requirement Gathering",
					customClass: "ganttRed",
				}]
			}, {
				name: " ",
				desc: "Scoping",
				values: [{
					from: "/Date(1322611200000)/",
					to: "/Date(1323302400000)/",
					label: "Scoping",
					customClass: "ganttRed"
				}]
			}, {
				name: "Sprint 1",
				desc: "Development",
				values: [{
					from: "/Date(1323802400000)/",
					to: "/Date(1325685200000)/",
					label: "Development",
					customClass: "ganttGreen"
				}]
			}, {
				name: " ",
				desc: "Showcasing",
				values: [{
					from: "/Date(1325685200000)/",
					to: "/Date(1325695200000)/",
					label: "Showcasing",
					customClass: "ganttBlue"
				}]
			}, {
				name: "Sprint 2",
				desc: "Development",
				values: [{
					from: "/Date(1326785200000)/",
					to: "/Date(1325785200000)/",
					label: "Development",
					customClass: "ganttGreen"
				}]
			}, {
				name: " ",
				desc: "Showcasing",
				values: [{
					from: "/Date(1328785200000)/",
					to: "/Date(1328905200000)/",
					label: "Showcasing",
					customClass: "ganttBlue"
				}]
			}, {
				name: "Release Stage",
				desc: "Training",
				values: [{
					from: "/Date(1330011200000)/",
					to: "/Date(1336611200000)/",
					label: "Training",
					customClass: "ganttOrange"
				}]
			}, {
				name: " ",
				desc: "Deployment",
				values: [{
					from: "/Date(1336611200000)/",
					to: "/Date(1336711200000)/",
					label: "Deployment",
					customClass: "ganttOrange"
				}]
			}],
			// _toSource(value),
			navigate: "scroll",
			scale: "weeks",
			minScale: "hours",
			maxScale: "months",
			itemsPerPage: 10,
			onItemClick: function(data) {
				alert("Item clicked - show some details");
			},
			onAddClick: function(dt, rowId) {
				alert("Empty space clicked - add an item!");
			}
		});

		// setTimeout(function() {
		//     var $$bars = $(".bar");
		//     // $$bars.popover({
		//     //     title: "I'm a popover",
		//     //     content: "And I'm the content of said popover.",
		//     //     trigger: "hover"
		//     // });
		//     $$bars.hover(function(event) {
		//         // event.preventDefault();
		//         // event.stopPropagation();
		//         // $(this).popover({
		//         //     title: "I'm a popover",
		//         //     content: $(this).text() + "... from gantt task"
		//         // }).popover("show");
		//         $(this).popover("title", $(this).text() + "... from gantt task");
		//     });
		// }, 100);
		this._initialized = true;
	}
};

exports.render = function() {
	var id = this._dataValue.getAttribute("id");
	this.$$widgetBox = $('<div style="height: 400px; border: 1px solid #aaa">').attr('id', "gtt_" + id).addClass('gtt-container').appendTo(this._dataValue);
};

function _toSource(value) {
	return {
		name: "Example",
		desc: "Lorem ipsum dolor sit amet.",
		values: value.$events && value.$events.map(function(e) {
			return {
				to: "/Date(" + (e.start && (e.start.date || e.start.datetime)) + ")/",
				from: "/Date(" + (e.end && (e.end.date || e.end.datetime)) + ")/",
				desc: e.description,
				label: e.summary,
				customClass: "ganttRed",
				dataObj: {}
			};
		})
	};
};
