// syracuse-ui/lib/page/landing/select/selectPage.js
"use strict";
var helpers = require("syracuse-core/lib/helpers");
var DesktopPage = require('syracuse-ui/lib/page/desktopPage').DesktopPage;
var FacetsBar = require('syracuse-ui/lib/search/facetsBar').FacetsBar;
var _browse = require('syracuse-ui/lib/page/landing/select/browseMenus');

var _gadgetTypes = ["requests", "processes", "stats", "representation", "calendar", "extLink"];

function _addBlock(slot, $title) {
	var block = syra_dom.addDiv("s-sv-wiz-step-block");
	var title = document.createElement("label");
	title.className = "s-sv-wiz-step-block-title";
	title.textContent = $title;
	block.appendChild(title);

	slot.appendChild(block);
	return {
		block: block,
		title: title,
		body: syra_dom.addDiv("s-sv-wiz-step-body", block)
	};
}

function _addStep1(page) {
	var step = syra_dom.addDiv("s-sv-wiz-step");
	var box = _addBlock(step, syra_local.ldpSelectEndpoint);
	var ldpPage = syra_site.landingPage;
	var isEndpointHidden = ldpPage && ldpPage.ldpRecord && ldpPage.ldpRecord.useCurrentEndpoint;
	syra_dom.hide(box.block, isEndpointHidden);

	page.loadNewItem(box.body, {
		$isTitleHidden: true,
		$bind: "endpoint"
	});


	var box = _addBlock(step, isEndpointHidden ? syra_local.ldpBrowseCategoryNoEndpoint : syra_local.ldpBrowseCategory);
	var rowCount = 0;
	for (var ii = 0, jj = _gadgetTypes.length; ii < jj; ii++) {
		page.loadNewItem(box.body, {
			$category: "link",
			$bind: _gadgetTypes[ii],
			$css: "s-sv-type-link s-sv-type-" + _gadgetTypes[ii],
			$skin: "s-sv-type-skin"
		});
		if (++rowCount == 3) {
			rowCount = 0;
			var sep = document.createElement("div");
			sep.className = "s-sv-wiz-cat-sep";
			box.body.appendChild(sep);
		}
	}
	page.layoutContent.domItem.appendChild(step);
}

function _addStep2(page) {
	var step = syra_dom.addDiv("s-sv-wiz-step");

	var box = _addBlock(step, syra_local.ldpUseSearch);
	page.loadNewItem(box.body, {
		$bind: "$searchVignetteField",
		$isTopLabelAlignment: false
	});
	page.searchVignetteBlock = _addBlock(step, syra_local.ldpNoVignetteFounds);
	page.searchVignetteBlock.block.style.display = "none";
	page.searchVignetteBlock.body.className += " s-sv-wiz-results";

	page.facetBar = new FacetsBar();
	page.facetBar.load(page, {
		slot: syra_dom.addDiv("s-sv-wiz-results-bar", page.searchVignetteBlock.body),
		$facetSkin: "s-search-vignette",
	}, {
		$filters: ["menuModule", "menuCategory"],
		$order: ["menuModule", "menuCategory"]
	});

	page.searchResultField = page.loadNewItem(syra_dom.addDiv("s-sv-wiz-results-data", page.searchVignetteBlock.body), {
		$bind: "$searchResult",
		$isEditMode: false,
		$isMenuRecordHidden: true,
		$noDataText: syra_local.search_no_res,
		$format: "cards",
		$alternateStyle: false,
		$skin: "s-sv-search-res",
		$selectMode: "multi",
		$selectByRow: true,
		renderRecordContent: function(record) {
			var title = document.createElement("div");
			title.innerHTML = syra_dom.formatHTMLText(record.dataset.title);
			title.className = "s-sv-search-res-title";
			record.body.appendChild(title);
			if (record.dataset.module) {
				var div = record.searchModuleItem = document.createElement("div");
				div.className = "s-sv-search-res-mod-value";
				div.innerHTML = syra_dom.formatHTMLText(record.dataset.module);
				record.body.appendChild(div);
			}
			if (record.dataset.categories) {
				record.searchCategoryItems = [];
				for (var ii = 0, jj = record.dataset.categories.length; ii < jj; ii++) {
					var div = document.createElement("div");
					div.className = "s-sv-search-res-cat-value";
					div.innerHTML = syra_dom.formatHTMLText(record.dataset.categories[ii]);
					record.body.appendChild(div);
					record.searchCategoryItems.push(div);
				}
			}
			if (page.findSelectedVignette(record.dataset.$key)) {
				syra_dom.toggleClass(record.domItem, "s-list-record-selected", true);
				_onSelectSearchRecord(record, true);
			}
		},
		$layout: {
			$items: []
		}
	});
	page.browseBlock = _addBlock(step, "");
	page.layoutContent.domItem.appendChild(step);
}

function _onSelectSearchRecord(record, selected) {
	record.searchModuleItem && syra_dom.toggleClass(record.searchModuleItem, "s-list-record-selected", selected);
	if (record.searchCategoryItems) {
		for (var ii = 0, jj = record.searchCategoryItems.length; ii < jj; ii++) {
			syra_dom.toggleClass(record.searchCategoryItems[ii], "s-list-record-selected", selected);
		}
	}
	record.$isSelected = selected;
}

function _gotoStep(page, index) {
	var steps = page.layoutContent.domItem.children;
	index--;
	for (var ii = 0, jj = steps.length; ii < jj; ii++) {
		steps[ii].style.display = (ii == index) ? "" : "none";
	}
}

function _updateSelectedEndpoint(page, endpoint) {
	page.selectedEndpoint = endpoint;
	syra_dom.empty(page.securityViewSlot);
	syra_menus.button.remove(page.securityRoleFlag);
	syra_menus.button.remove(page.securityEndpointFlag);
	if (endpoint) {
		page.securityEndpointFlag = syra_menus.button.add({
			isIndicator: true,
			parent: page,
			slot: page.securityViewSlot,
			text: endpoint.$value || endpoint.description,
			css: "s-security-user",
			fontIcon: "endpoint"
		});
	}
}

function _applySearchChange(page, data) {
	page.startChange();
	page.browseBlock.block.style.display = "none";
	page.searchVignetteBlock.block.style.display = "";
	data.$searchResult = data.$resources;
	delete data.$resources;
	page.applyChange({
		$searchResult: null
	});
	page.applyChange(data);
	page.endChange();
}

function _updatePage(page, data) {
	page.startChange();
	page.applyChange({
		$searchVignetteField: "",
		selectedEndpoint: page.selectedEndpoint
	});
	syra_dom.empty(page.browseBlock.body);
	_browse.buildLevels(page, page.browseBlock.body, data);
	page.searchVignetteBlock.block.style.display = "none";
	page.browseBlock.block.style.display = "";
	page.endChange();
}

function _callServer(page, method, url, onSuccess, options) {
	syra_controller.callServer(page, options || {
		method: method || "GET",
		$location: {
			$url: url
		}
	}, onSuccess, function(error) {
		var $diagnoses;
		if (error.data.indexOf("$diagnoses") != -1) {
			$diagnoses = JSON.parse(error.data).$diagnoses;
		} else {
			$diagnoses = [{
				$severity: "error",
				$message: error.data
			}];
		}
		syra_diagnose.showDiagnoses({
			$diagnoses: $diagnoses
		});
	});
}

function SelectPage() {}

exports.SelectPage = helpers.defineClass(SelectPage, DesktopPage, {
	onItemInOut: function(onEnter, event, target) {
		if (target.className.indexOf("s-sv-browse-item") >= 0) {
			syra_dom.toggleClass(target, "s-record-enter", onEnter);
		} else {
			this.facetBar.onItemInOut(onEnter, event, target);
			DesktopPage.prototype.onItemInOut.call(this, onEnter, event, target);
		}
	},

	onSelectRecord: function(records, record, isSelected) {
		if (records) {
			this.selectVignette(record.dataset, isSelected);
			_onSelectSearchRecord(record, isSelected);
		}
	},
	fetch: function(options) {
		var self = this;
		if (options && options.isPageLoading) {
			options.method = "POST";
			DesktopPage.prototype.fetch.call(self, options);
		} else {
			options.method = options.method || "GET";
			options.$location = options.$location || {
				$url: self.requestUrlComplete
			};
			_callServer(self, undefined, undefined, function(data) {
				_applySearchChange(self, data);
			}, options);
		}
	},
	getPrototype: function() {
		this.selectedVignettes = []; // adding prototype properties for search field and results page
		if (this.$prototype && this.$prototype.$properties) {
			this.$prototype.$properties.$searchVignetteField = {
				$type: "application/x-string",
				$title: syra_local.ldpSearch
			};
			this.$prototype.$properties.$searchResult = {
				$type: "application/x-array",
				$item: {
					$type: "application/json",
					$properties: {}
				}
			};
		};
		delete this.$prototype.$description; // in order not to display description text value
		return this.$prototype;
	},
	onMenuClick: function(menu) {
		var self = this;
		if (_gadgetTypes.indexOf(menu.$bind) >= 0) {
			self.browseBlock.title.title = self.browseBlock.title.textContent = menu.getTitle();
			self.requestUrl = menu.$searchUrl || menu.$url;
			_callServer(self, menu.$method, menu.$url, function(data) {
				syra_menus.button.hide(self.backBtn, false);
				_gotoStep(self, 2);
				_updatePage(self, data);
			});
			return false;
		}
		return false;
	},
	selectVignette: function(dataVignette, selected) {
		var selected = selected || false;
		var vignette = this.findSelectedVignette(dataVignette.$key || dataVignette.$uuid);
		if (vignette) {
			!selected && this.selectedVignettes.splice(this.selectedVignettes.indexOf(vignette), 1);
		} else {
			selected = true;
			this.selectedVignettes.push({
				vignette: dataVignette,
				endpoint: {
					$uuid: this.selectedEndpoint.$uuid
				}
			});
		}
		this.dialogWrapper.hideOkButton(this.selectedVignettes.length <= 0);
		return selected;
	},
	findSelectedVignette: function($uuid) {
		for (var ii = 0, jj = this.selectedVignettes.length; ii < jj; ii++) {
			var sel = this.selectedVignettes[ii];
			if ((sel.vignette.$key || sel.vignette.$uuid) == $uuid) {
				return sel;
			}
		}
		return null;
	},
	onNotifyDataChange: function(field, value) {
		var self = this;
		switch (field.$item.$bind) {
			case "$searchVignetteField":
				if (value != '') {
					// change count value
					self.requestUrlComplete = self.requestUrl.replace(/(count=)\d+(.+)/g, "$120$2") + ("&search=" + value);
					_callServer(self, "GET", self.requestUrlComplete, function(data) {
						_applySearchChange(self, data, true);
					});
				} else {
					self.endChange();
				}
				return false;
			case "endpoint":
				_updateSelectedEndpoint(self, value);
				return true;
			default:
				return true;
		}
	},
	applyChange: function(newData) {
		if (newData) {
			newData.endpoint && _updateSelectedEndpoint(this, newData.endpoint);
		}
		DesktopPage.prototype.applyChange.call(this, newData);
		newData && newData.$searchFacets && this.facetBar.fill(this.dataset.$searchFacets);
	},
	ensureDefaultArticle: function($article, $prototype) {
		$prototype.$title = syra_local.ldpAddNewVignette;
		return {
			$layout: {
				$items: []
			}
		};
	},
	drawBox: function() {
		this.isMenuBarDisabled = true;
		DesktopPage.prototype.drawBox.call(this);
		_addStep1(this);
		_addStep2(this);
		this.backBtn = syra_menus.button.add({
			parent: this,
			text: syra_local.ldpBack,
			css: "s-sv-back-link",
			iconOnly: true,
			fontIcon: "reload_back",
			isHidden: true,
			btnclick: function() {
				syra_menus.button.hide(this, true);
				_gotoStep(this.parent, 1);
				this.parent.resizeArticle();
			}
		});
		this.domTitle.parentNode.insertBefore(this.backBtn.link, this.domTitle);
		_gotoStep(this, 1);
	},
	onFacetFilterChange: function(options) {
		this.jsonParams = options.jsonParams;
		options.method = "GET";
		this.fetch(options);
	},
	dispose: function() {
		_browse.dispose();
		syra_site.disposeObject(this.facetBar);
		DesktopPage.prototype.dispose.call(this);
	}
});
