// syracuse-ui/lib/links/navigation/navigator.js
"use strict";
var helpers = require('syracuse-core').helpers;
var DesktopPage = require('syracuse-ui/lib/page/desktopPage').DesktopPage;
var _module = require('./module');

var _mode = {
	mega: require('./mega/mega'),
	map: require('./sitemap/sitemap'),
	edit: require('./edit/edit'),
	breadcrumb: require('./breadcrumb/breadcrumb')
};
exports.megaMenus = null;

function Navigator() {}

helpers.defineClass(Navigator, DesktopPage, {
	load: function(initData) {
		var activePage = syra_site.getActivePage();
		this.isActivePageFusion = activePage && activePage.isFusionPage;

		(this.page = this).$item = {
			$layout: {
				$items: []
			}
		};

		syra_pageBuilder.initialize(this);
		this.externalAdapter = syra_site.externalAdapter;
		syra_article.beforeDraw(this);
		this.mode.page.draw && this.mode.page.draw(this, initData);
		syra_article.endDraw(this);

		this.addChildren();
		this.areModulesLoaded = true;
	},

	addChildren: function() {
		var modules = this.dataset && this.dataset.modules;
		this.children = [];
		if (modules) {
			for (var ii = 0, jj = modules.length; ii < jj; ii++) {
				this.addChild(modules[ii]);
			}
		}
		this.ensureNoChild();
	},
	ensureNoChild: function() {
		if (this.children.length) {
			if (this.noChild) {
				syra_dom.remove(this.noChild);
				delete this.noChild;
			}

		} else {
			syra_dom.empty(this.modulesBar);
			(this.noChild = syra_dom.li(this.page.$skin + "-no-modules", this.modulesBar)).textContent = syra_local.nvpNoModule;
		}
	},
	addChild: function(record) {
		var child = _module.add(this, record);
		this.children.push(child);
		return child;
	},
	clear: function() {
		syra_dom.empty(this.mega_popup);
		syra_dom.empty(this.modulesSlot);
		this.more && this.more.dispose();
		this.more = this.lastSelected = null;
		if (this.children) {
			for (var ii = 0, jj = this.children.length; ii < jj; ii++) {
				this.disposeChild(this.children[ii]);
			}
		}
	},
	disposeChild: function(child) {
		if (child) {
			if (child.children) {
				for (var ii = 0, jj = child.children.length; ii < jj; ii++) {
					this.disposeChild(child.children[ii]);
				}
			}
			child.carousel && child.carousel.dispose();
			syra_item.unregister(child);
			syra_site.disposeObject(child);
		}
	},
	dispose: function() {
		this.clear();
		this.mode && this.mode.page.dispose(this);
		DesktopPage.prototype.dispose.call(this);
	}
});


exports.toggleMegaMenus = function(show, $representation) {
	var isShown = syra_dom.isDisplay(syra_site.megaMenusBar);
	if (show) {
		if (exports.megaMenus) {
			if ((syra_site.userProfile.selectedEndpointUiid !== exports.megaMenus.selectedEndpointUiid) || (syra_site.userProfile.selectedRoleUiid !== exports.megaMenus.selectedRoleUiid)) {
				exports.megaMenus.dispose();
				exports.megaMenus = null;
			}
		}

		if (!exports.megaMenus) {
			var page = exports.megaMenus = new Navigator();
			exports.megaMenus.selectedEndpointUiid = syra_site.userProfile.selectedEndpointUiid;
			exports.megaMenus.selectedRoleUiid = syra_site.userProfile.selectedRoleUiid;
			page.$representation = $representation;
			(page.mode = _mode.mega).page.load(page);
		}
	}
	if (isShown != show) {
		syra_dom.hide(syra_site.megaMenusBar, !show);
		syra_site.resizeItem();
	}

};
exports.showSiteMap = function() {
	syra_bookmarks.open(function() {
		var page = new Navigator();
		page.mode = _mode.map;
		page.$prototype = exports.megaMenus.$prototype;
		page.dataset = exports.megaMenus.dataset;
		page.load();
		if (page.children.length) {
			page.children[0].activate(true);
		}
		page.isMapLoaded = true;
		page.resizeItem();
	});
};

exports.editSiteMap = function() {
	var page = new Navigator();
	page.mode = _mode.edit;
	return page;
};

exports.showBreadcrumbs = function(breadcrumbs, options) {
	var page = syra_site.breadcrumbs = new Navigator();
	page.breadcrumbs = breadcrumbs;
	page.breadcrumbIndex = options.menu.index;

	(page.mode = _mode.breadcrumb).page.load(page, options);

	if (options.menu.index === 0) {
		page.mode.modules.activate(true, page);
	} else {
		var menuNode = page.mode.modules.getModule(breadcrumbs, options.menu.index, syra_site.breadcrumbs.children);

		switch (options.menu.index) {
			case 1:
				page.mode.module.activate(true, menuNode);
				break;
			case 2:
				page.mode.subModule.activate(true, null, menuNode);
				break;
			default:
				page.mode.menuBlock.activate(true, menuNode);
		}
	}
};
