// @sage/syracuse-react/src/__tests__/testHelper.ts
"use strict";

const enableFakeSessionStorage = () => {
    global.sessionStorage = {
        getItem: function (key: any) {
            return this[key];
        },
        setItem: function (key: any, value: any) {
            this[key] = value;
        },
        removeItem: function (key: any) {
            delete this[key];
        }
    };
}

const enableFakeLocalStorage = () => {
    global.localStorage = {
        getItem: function (key: any) {
            return this[key];
        },
        setItem: function (key: any, value: any) {
            this[key] = value;
        },
        removeItem: function (key: any) {
            delete this[key];
        }
    };
}

const BODY_ID = "s_app_body";
const SLOT_ID = "s_app_slot";

process.env.NODE_ENV = 'test';

global.syra_core = require('./mock/syra_core.json');
global.syra_menus = {
    click: {
        fire: jest.fn()
    }
}

enableFakeLocalStorage();
enableFakeSessionStorage();

//console.log(global.localStorage);

import { Bookmarks } from './../navigation/bookmark/bookmarks';
import { Sitemap } from './../navigation/sitemap/sitemap';
import { Profile } from './../profile/profile';
import { IEndpoint } from './../profile/endpoint';
import { SitemapModule } from './../navigation/sitemap/groups';
import { OVER_ROOT_ID } from './../over/overHelper';

import Store from './../dataStore/store';

class Fixtures {

    private loadedMockFiles: any = [];

    private _loadMockFile(filename: string): any {
        if (this.loadedMockFiles.hasOwnProperty(filename)) {
            return JSON.parse(JSON.stringify(this.loadedMockFiles[filename]));
        } else {
            let loadedFileContent = require('./mock/' + filename + '.json');
            this.loadedMockFiles[filename] = JSON.parse(JSON.stringify(loadedFileContent));
            return loadedFileContent;
        }
    }

    getBookmarks(profileEndpoint: IEndpoint, dataFilename: string = 'bookmark_data'): Bookmarks {
        return new Bookmarks(this._loadMockFile('data/' + dataFilename), profileEndpoint);
    }

    getProfile(dataFilename: string = 'profile_data'): Profile {
        return new Profile(this._loadMockFile('prototype/profile_prototype'), this._loadMockFile('data/' + dataFilename));
    }

    getProfilePopulated(dataFilename: string = 'profile_data'): Profile {
        let profile = this.getProfile(dataFilename);
        let sitemap = this.getSitemap();
        let store = this.getStoreWithDataset('sitemap_data');

        let bookmarks = this.getBookmarks(profile.selectedEndpoint);

        let modules: SitemapModule[] = [];
        if (store.dataset.modules) {
            for (let mod of store.dataset.modules) {
                modules.push(new SitemapModule(sitemap, mod));
            }
            modules[0].selected = true;
        }
        profile.sitemap = sitemap;
        profile.bookmarks = bookmarks;
        profile.dataset.user = { firstName: 'James' };

        return profile;
    }

    getSitemap(dataFilename: string = 'sitemap_data'): Sitemap {
        return new Sitemap(this._loadMockFile('prototype/sitemap_prototype'), this._loadMockFile('data/' + dataFilename));
    }

    getStoreWithDataset(dataFilename: string): Store {
        return new Store(null, this._loadMockFile('data/' + dataFilename));
    }
}

class TestHelper {

    appHostNode: HTMLDivElement;
    fixtures: Fixtures = new Fixtures();

    constructor() {
        this.addAppSlotNode();
    }

    addAppSlotNode(): HTMLDivElement {
        if (!this.appHostNode) {
            this.appHostNode = document.createElement('div');
            this.appHostNode.setAttribute('id', SLOT_ID);
            document.body.appendChild(this.appHostNode);

            const overContainer = document.createElement('div');
            overContainer.setAttribute('id', OVER_ROOT_ID);
            this.appHostNode.appendChild(overContainer);
        }

        return this.appHostNode;
    }

    removeAppSlotNode() {
        document.removeChild(this.appHostNode);
        this.appHostNode = null;
    }

    createFakeLocalStorage() {
        if (!global.localStorage) {
            enableFakeLocalStorage();
        }
    }
    deleteFakeLocalStorage() {
        delete global.localStorage;
    }

    createFakeSessionStorage() {
        if (!global.sessionStorage) {
            enableFakeSessionStorage();
        }
    }

    deleteFakeSessionStorage() {
        delete global.sessionStorage;
    }
}

export default new TestHelper();
