// test-contract/lib/contract.js
"use strict"

exports.contract = {
	application : "example",
	contract : "admin",
	entities : {
		userTest : require('./entities/userTest').entity,
		userInfo : require('./entities/userInfo').entity,
		country : require('./entities/country').entity,
		post : require("./entities/post").entity,
		address : require('./entities/address').entity,
		refA : require('./entities/refA').entity,
		refB : require('./entities/refB').entity,
		refC : require('./entities/refC').entity,
		storeTest : require('./entities/storeTest').entity,
		propagateAndValidation : require('./entities/propagateAndValidation').entity,
		typesTest : {
			$properties : {
				bool : {
					$type : "boolean"
				}
			}
		},
		refD : require('./entities/refD').entity,
		signedTest : {
			$signed : [ "description", "description2" ],
			$properties : {
				description : {},
				description2 : {}
			}
		},
		localized : {
			$properties : {
				description : {
					$isLocalized : true
				}
			}
		},
		caseInsensitive : {
			$properties : {
				ci : {
					$caseInsensitive : true,
					$isUnique : true
				},
				cs : {}
			}
		},
		polymorph: {
			$properties: {
				description: {}
			},
			$relations: {
				poly: {
					$variants: {
						A: {
							$type: "refA"
						},
						B: {
							$type: "refB",
							$isChild: true
						},
						C: {
							$type: "refC"
						}
					}
				},
				pluralPoly: {
					$isPlural: true,
					$variants: {
						A: {
							$type: "refA"
						},
						B: {
							$type: "refB",
							$isChild: true
						}
					}
				}
			}
		}
	}
};
