// syracuse-ui/lib/authoring/setting/paletteSettings.js
"use strict";
function getRow(isField){
    return {
        "$category": "section",
        "$skin": "s-aw-h2",
        "$css": "s-aw-palette-layout",
        "$isBoxCollapsable": true,
        "$title": "{@rowLayout}",
        "$layout": {
            "$items": [{
                $category: "menus",
                $title: "{@addrow}",
                $skin: "s-aw-cols",
                $isHidden: true,
                "$itemSkin": "s-aw-cols-link",
                "$layout": {
                    "$layoutType": "stack",
                    "$items": [{
                        "$css": "s-aw-cols-add-before",
                        "$bind": "$addRowBefore"
                    }, {
                        "$css": "s-aw-cols-add-after",
                        "$bind": "$addRowAfter"
                    }]
                }
            }, {
                $category: "menus",
                $title: "{@columns}",
                $skin: "s-aw-cols",
                "$itemSkin": "s-aw-cols-link",
                "$layout": {
                    "$layoutType": "row",
                    "$fitToContent": true,
                    "$items": [{
                        "$bind": "s-layout-type-1"
                    }, {
                        "$bind": "s-layout-type-2"
                    }, {
                        "$bind": "s-layout-type-3"
                    }, {
                        "$bind": "s-layout-type-4"
                    }, {
                        "$bind": "s-layout-type-5"
                    }, {
                        "$bind": "s-layout-type-6"
                    }]
                }
            }]
        }
    };
}

function getLayout($level){
    var $layout = {
        "$category": "section",
        "$skin": "s-aw-h2",
        "$css": "s-aw-palette-" + $level,
        "$isBoxCollapsable": true,
        "$title": "{@layout}",
        "$layout": {
            "$items": [{
                $category: "menus",
                $title: "{@columns}",
                $skin: "s-aw-cols",
                "$itemSkin": "s-aw-cols-link",
                "$layout": {
                    "$layoutType": "row",
                    "$fitToContent": true,
                    "$items": [{
                        "$bind": "s-layout-type-1",
                        "$css": $level == "field" ? "" : "s-aw-row-box"
                    }, {
                        "$bind": "s-layout-type-2",
                        "$css": $level == "field" ? "" : "s-aw-row-box"
                    }, {
                        "$bind": "s-layout-type-3",
                        "$css": $level == "field" ? "" : "s-aw-row-box"
                    }, {
                        "$bind": "s-layout-type-4",
                        "$css": $level == "field" ? "" : "s-aw-row-box"
                    }, {
                        "$bind": "s-layout-type-5",
                        "$css": $level == "field" ? "" : "s-aw-row-box"
                    }, {
                        "$bind": "s-layout-type-6",
                        "$css": $level == "field" ? "" : "s-aw-row-box"
                    }]
                }
            }, {
                "$category": "field",
                "$isHidden": !($level == "article" || $level == "section"),
                "$css": "s-layout-tab-choice",
                "$bind": "$layoutTabChoice",
                "$skin": "s-aw-field",
                "$isEditMode": true,
                "$format": "$button",
                "$choiceLayout": "row",
                "$icon": {
                    "$mode": "icon",
                    "$path": "authoring/s-layout-"
                }
            }]
        }
    };
    if ($level == "article") {
        $layout.$layout.$items.push({
            $title: "{@models}",
            $id: "s-awModels",
            $isHidden: false,
            "$category": "menus",
            "$itemIcon": {
                "$mode": "icon",
                "$path": "authoring/s-aw-"
            },
            "$skin": "s-aw-menus",
            "$itemSkin": "s-aw-menus-link",
            "$isBoxCollapsable": "true",
            "$opened": false,
            "$layout": {
                "$items": [{
                    "$layoutType": "row",
                    "$items": [{
                        "$icon": null,
                        "$bind": "pattern-headerTabs"
                    }, {
                        "$icon": null,
                        "$bind": "$defaultLayout"
                    }]
                }, {
                    "$layoutType": "row",
                    "$items": [{
                        "$bind": "pattern-1"
                    }, {
                        "$bind": "pattern-2"
                    }, {
                        "$bind": "pattern-3"
                    }, {
                        "$bind": "pattern-4"
                    }]
                }]
            }
        });
    }
    return $layout;
}

exports.$item = {
    $menus: {
        $skin: "s-aw-menus",
        $layout: {
            $layoutType: "row",
            $items: [{
                "$category": "menus",
                "$skin": "s-aw-menus",
                "$itemSkin": "s-aw-menus-link",
                "$itemIcon": {
                    "$mode": "icon",
                    "$path": "authoring/s-aw-"
                },
                $layout: {
                    $layoutType: "row",
                    $fitToContent: true,
                    $items: [{
                        "$bind": "$help"
                    }, {
                        "$bind": "$previousLayout"
                    }, {
                        "$bind": "$nextLayout"
                    }, {
                        $icon: null,
                        "$bind": "$undoAllLayout"
                    }]
                }
            }, {
                $category: "field",
                "$bind": "$viewType",
                "$skin": "s-aw-viewType",
                "$isEditMode": true,
                "$format": "$button",
                "$choiceLayout": "row"
            }, {
                "$title": "{@insert}",
                "$category": "menus",
                $isHidden: false,
                "$itemIcon": {
                    "$mode": "iconText",
                    "$path": "authoring/s-aw-"
                },
                "$skin": "s-aw-menus",
                "$itemSkin": "s-aw-menus-link",
                "$isBoxCollapsable": "true",
                "$isPopupContent": true,
                $css: "s-aw-insert",
                $layout: {
                    $items: [{
                        $layoutType: "row",
                        $items: [{
                            "$css": "s-aw-add-section",
                            $bind: "$addSection",
                            $value: "add-section"
                        }, {
                            "$css": "s-aw-add-block",
                            $bind: "$addBlock",
                            $value: "add-block"
                        }]
                    }, {
                        "$title": "{@addField}",
                        "$id": "$awAddField",
                        "$category": "menus",
                        "$itemSkin": "s-aw-mn-field-link",
                        "$isBoxCollapsable": false,
                        $css: "s-aw-insert",
                        "$titleIcon": {
                            "$value": "template",
                            "$path": "authoring/s-aw-"
                        },
                        "$itemIcon": {
                            "$mode": "iconText",
                            "$path": "authoring/s-aw-"
                        },
                        $layout: {
                            $layoutType: "row",
                            $items: [{
                                $layoutType: "stack",
                                $css: "s-aw-insert-col",
                                "$items": []
                            }, {
                                $layoutType: "stack",
                                $css: "s-aw-insert-col",
                                "$items": []
                            }, {
                                $layoutType: "stack",
                                $css: "s-aw-insert-col",
                                "$items": []
                            }]
                        }
                    }]
                }
            }]
        }
    },
    $isTitleHidden: true,
    $skin: "s-aw-palette-item",
    $layout: {
        $items: [{
            "$paletteId": "article",
            "$category": "section",
            "$skin": "s-aw-h1",
            "$css": "s-aw-palette-article",
            "$title": "{@page}",
            "$isHidden": true,
            "$layout": {
                "$items": [{
                    "$category": "section",
                    "$isTitleHidden": true,
                    "$skin": "s-aw-palette-item",
                    "$layout": {
                        "$layoutType": "stack",
                        "$items": [getRow(), getLayout("article"), {
                            "$category": "section",
                            "$skin": "s-aw-h2",
                            "$css": "s-aw-palette-article",
                            "$bind": "$field",
                            "$isBoxCollapsable": true,
                            "$title": "{@sectionFields}",
                            "$layout": {
                                "$layoutType": "stack",
                                "$items": [{
                                    "$bind": "$fieldsIsTitleHidden",
                                    "$skin": "s-aw-field",
                                    "$isEditMode": true,
                                    "$isLeftSpaceHidden": true
                                }, {
                                    "$bind": "$labelAlignment",
                                    "$skin": "s-aw-field",
                                    "$isEditMode": true,
                                    "$isLeftSpaceHidden": true,
                                    "$choiceLayout": "1"
                                }]
                            }
                        }]
                    }
                }]
            }
        }, {
            "$paletteId": "section",
            "$category": "section",
            "$skin": "s-aw-h1",
            "$css": "s-aw-palette-section",
            "$title": "{@section}",
            "$isHidden": true,
            "$layout": {
                "$items": [{
                    "$category": "section",
                    "$id": "s-aw-palette-section",
                    "$isTitleHidden": true,
                    "$skin": "s-aw-palette-item",
                    "$layout": {
                        "$layoutType": "stack",
                        "$items": [getRow(), getLayout("section"), {
                            "$category": "section",
                            "$skin": "s-aw-h2",
                            "$css": "s-aw-palette-section",
                            "$isBoxCollapsable": true,
                            "$title": "{@sectionAppearance}",
                            "$layout": {
                                "$items": [{
                                    "$bind": "$excludeMe",
                                    "$category": "link",
                                    "$skin": "s-aw-menus-link",
                                    "$css": "s-aw-exclude-me"
                                }, {
                                    "$bind": "$title",
                                    "$skin": "s-aw-field",
                                    "$isEditMode": true,
                                    "$isLeftSpaceHidden": true
                                }, {
                                    "$bind": "$isTitleHidden",
                                    "$css": "s-aw-field-is-title-hidden",
                                    "$skin": "s-aw-field",
                                    "$isEditMode": true
                                }]
                            }
                        }, {
                            "$category": "section",
                            "$skin": "s-aw-h2",
                            "$css": "s-aw-palette-section",
                            "$isBoxCollapsable": true,
                            "$title": "{@sectionBehaviour}",
                            "$layout": {
                                "$items": [{
                                    "$bind": "$isBoxCollapsable",
                                    "$skin": "s-aw-field",
                                    "$isEditMode": true,
                                    "$isLeftSpaceHidden": true
                                }, {
                                    "$bind": "$isMaximizable",
                                    "$skin": "s-aw-field",
                                    "$isEditMode": true,
                                    "$isLeftSpaceHidden": true
                                }]
                            }
                        }, {
                            "$category": "section",
                            "$skin": "s-aw-h2",
                            "$css": "s-aw-palette-section",
                            "$isBoxCollapsable": true,
                            "$bind": "$field",
                            "$title": "{@sectionFields}",
                            "$layout": {
                                "$layoutType": "stack",
                                "$items": [{
                                    "$bind": "$fieldsIsTitleHidden",
                                    "$skin": "s-aw-field",
                                    "$isEditMode": true,
                                    "$isLeftSpaceHidden": true
                                }, {
                                    "$bind": "$labelAlignment",
                                    "$skin": "s-aw-field",
                                    "$isEditMode": true,
                                    "$isLeftSpaceHidden": true,
                                    "$choiceLayout": "1"
                                }]
                            }
                        }]
                    }
                }]
            }
        }, {
            "$paletteId": "block",
            "$category": "section",
            "$skin": "s-aw-h1",
            "$css": "s-aw-palette-block",
            "$title": "{@block}",
            "$isHidden": true,
            "$layout": {
                "$items": [{
                    "$category": "section",
                    "$id": "s-aw-palette-block",
                    "$isTitleHidden": true,
                    "$skin": "s-aw-palette-item",
                    "$layout": {
                        "$layoutType": "stack",
                        "$items": [getRow(), getLayout("block"), {
                            "$category": "section",
                            "$skin": "s-aw-h2",
                            "$css": "s-aw-palette-block",
                            "$isBoxCollapsable": true,
                            "$title": "{@sectionAppearance}",
                            "$layout": {
                                "$items": [{
                                    "$bind": "$excludeMe",
                                    "$category": "link",
                                    "$skin": "s-aw-menus-link",
                                    "$css": "s-aw-exclude-me"
                                }, {
                                    "$bind": "$title",
                                    "$skin": "s-aw-field",
                                    "$isEditMode": true,
                                    "$isLeftSpaceHidden": true
                                }, {
                                    "$bind": "$isTitleHidden",
                                    "$css": "s-aw-field-is-title-hidden",
                                    "$skin": "s-aw-field",
                                    "$isEditMode": true
                                }]
                            }
                        }, {
                            "$category": "section",
                            "$skin": "s-aw-h2",
                            "$css": "s-aw-palette-block",
                            "$isBoxCollapsable": true,
                            "$title": "{@sectionBehaviour}",
                            "$layout": {
                                "$items": [{
                                    "$bind": "$isBoxCollapsable",
                                    "$skin": "s-aw-field",
                                    "$isEditMode": true,
                                    "$isLeftSpaceHidden": true
                                }, {
                                    "$bind": "$isMaximizable",
                                    "$skin": "s-aw-field",
                                    "$isEditMode": true,
                                    "$isLeftSpaceHidden": true
                                }]
                            }
                        }, {
                            "$category": "section",
                            "$skin": "s-aw-h2",
                            "$css": "s-aw-palette-block",
                            "$isBoxCollapsable": true,
                            "$bind": "$field",
                            "$title": "{@sectionFields}",
                            "$layout": {
                                "$layoutType": "stack",
                                "$items": [{
                                    "$bind": "$fieldsIsTitleHidden",
                                    "$skin": "s-aw-field",
                                    "$isEditMode": true,
                                    "$isLeftSpaceHidden": true
                                }, {
                                    "$bind": "$labelAlignment",
                                    "$skin": "s-aw-field",
                                    "$isEditMode": true,
                                    "$isLeftSpaceHidden": true,
                                    "$choiceLayout": "1"
                                }]
                            }
                        }]
                    }
                }]
            }
        }, {
            "$paletteId": "field",
            "$category": "section",
            "$skin": "s-aw-h1",
            "$css": "s-aw-palette-field",
            "$title": "{@field}",
            "$isHidden": true,
            "$layout": {
                "$items": [{
                    "$category": "section",
                    "$id": "s-aw-palette-field",
                    "$isTitleHidden": true,
                    "$skin": "s-aw-palette-item",
                    "$layout": {
                        "$layoutType": "stack",
                        "$items": [getRow(), getLayout("field"), {
                            "$category": "section",
                            "$css": "s-aw-palette-field",
                            "$skin": "s-aw-h2",
                            "$isBoxCollapsable": true,
                            "$title": "{@sectionAppearance}",
                            "$layout": {
                                "$layoutType": "stack",
                                "$items": [{
                                    "$bind": "$excludeMe",
                                    "$category": "link",
                                    "$skin": "s-aw-menus-link",
                                    "$css": "s-aw-exclude-me"
                                }, {
                                    "$bind": "$isTitleHidden",
                                    "$title": "{@isLabelHidden}",
                                    "$skin": "s-aw-field",
                                    "$isEditMode": true,
                                    "$isLeftSpaceHidden": true
                                }]
                            }
                        }, {
                            "$category": "section",
                            "$css": "s-aw-palette-field",
                            "$skin": "s-aw-h2",
                            "$isBoxCollapsable": true,
                            "$title": "{@sectionFieldLabel}",
                            "$layout": {
                                "$layoutType": "stack",
                                "$items": [{
                                    "$bind": "$labelAlignment",
                                    "$title": "{@sectionFieldLabelPosition}",
                                    "$skin": "s-aw-field",
                                    "$isEditMode": true,
                                    "$choiceLayout": "1"
                                }]
                            }
                        }, {
                            "$category": "section",
                            "$title": "{@sectionFieldWidget}",
                            "$isBoxCollapsable": true,
                            "$css": "s-aw-palette-field",
                            "$skin": "s-aw-h2",
                            "$isHidden": true,
                            "$id": "s-aw-field-widget",
                            "$layout": {
                                "$items": []
                            }
                        }]
                    }
                }]
            }
        }]
    }
};
