// syracuse-ui/lib/tests/pages/devPage.js
"use strict";
var helpers = require('syracuse-core').helpers;
var DesktopPage = require("syracuse-ui/lib/page/desktopPage").DesktopPage;

function DevPage() {}

exports.DevPage = helpers.defineClass(DevPage, DesktopPage, {
	addBox: function($title) {
		var box = {
			slot: syra_dom.div("s_section"),
			body: syra_dom.div("s_section_body"),
		};
		if ($title) {
			box.head = syra_dom.div("s_section_header", box.slot);
			syra_dom.text("s_section_title", $title, box.head);
		}
		box.slot.appendChild(box.body);
		return box;
	},
	load: function(initData) {
		var self = this;
		self.isAutoInsertFieldDisabled = true;
		self.isMenuBarDisabled = true;
		syra_pageBuilder.initialize(self);
		self.$prototype.$properties = self.$prototype.$properties || {};
		DesktopPage.prototype.load.call(self, initData);
		require.async("syracuse-ui/lib/tests/pages/" + self.$prototype.$devName, function(err, module) {
			if (err && err.message) {
				syra_alert.error(err.message);
			} else {
				self.devModule = module.map;
				var $links = {};
				self.devModule.$items.forEach(function($block, blockIndex) {
					var box = self.addBox($block.$title);
					self.layoutContent.domItem.appendChild(box.slot);
					box.cols = [];
					for (var mm = 0; mm < 4; mm++) {
						box.cols.push(syra_dom.div("col-lg-3", box.body));
					}
					var colIndex = 0;
					$block.$menus.forEach(function($menu, menuIndex) {
						var id = self.id + "-link-" + blockIndex + "-" + menuIndex;
						$links[id] = $menu;
						self.addItem(box.cols[colIndex++], {
							$category: "link",
							$skin: "s-sitemap-mega-item-btn-menu",
							$bind: id
						});
						if (colIndex == 4) {
							colIndex = 0;
						}
					});
				});
				self.applyChange({
					$title: self.devModule.$title,
					$links: $links
				});
				self.resizeItem(true);
			}
		});

	}
});
