// syracuse-ui/lib/tests/pages/map.js
"use strict";
exports.map = [{
	$title: "Array",
	$menus: [{
		$url: "?representation=s-uitest.$test&name=array_array&category=grid",
		$title: "Array Grid",
		$hasUnitTest: true
	}, {
		$url: "?representation=s-uitest.$test&name=array_array&category=gridcards",
		$title: "Array Grid - Cards",
		$hasUnitTest: true
	}, {
		$url: "?representation=s-uitest.$test&name=array_array&category=cards",
		$title: "Array Cards",
		$hasUnitTest: true
	}, {
		$url: "?representation=s-uitest.$test&name=array_array&category=selection",
		$title: "Array Selection",
		$hasUnitTest: true
	}, {
		$url: "?representation=s-uitest.$test&name=array_graph",
		$title: "Array Graph"
	}, {
		$url: "?representation=s-uitest.$test&name=array_tree",
		$title: "Array Tree "
	}, {
		$url: "?representation=s-uitest.$test&name=array_single",
		$title: "Array Single"
	}]

}, {
	$title: "Fields",
	$menus: [{
		$url: "?representation=s-uitest.$test&name=field_field&category=simple",
		$title: "Fields",
		$hasUnitTest: true
	}, {
		$url: "?representation=s-uitest.$test&name=field_field&category=binary",
		$title: "Fields Binary",
		$hasUnitTest: true
	}, {
		$url: "?representation=s-uitest.$test&name=field_field&category=editor",
		$title: "Fields Editor"
	}, {
		$url: "?representation=s-uitest.$test&name=field_child",
		$title: "Fields Child Object"
	}, {
		$url: "?representation=s-uitest.$test&name=field_field&category=icon",
		$title: "Fields Icon",
		sds$hasUnitTest: true
	}]
}, {
	$title: "Misc",
	$menus: [{
		$url: "?representation=s-uitest.$test&name=navigation_megamenu",
		$title: "navig",
	}, {
		$url: "?representation=s-uitest.$test&name=misc_office",
		$title: "office",
	}, {
		$url: "?representation=s-uitest.$test&name=misc_dialogsTools",
		$title: "Dialogs Tools"
	}, {
		$url: "?representation=s-uitest.$test&name=process_process",
		$title: "Process"
	}, {
		$url: "?representation=s-uitest.$test&name=misc_alert",
		$title: "Diagnoses"
	}, {
		$url: "?representation=s-uitest.$test&name=trackers_trackers",
		$title: "Trackers"
	}, {
		$url: "?representation=s-uitest.$test&name=misc_parameters",
		$title: "Dynamic Parameters"
	}, {
		$url: "?representation=s-uitest.$test&name=misc_layout",
		$title: "Layout"
	}]
}, {
	$title: "extends",
	$menus: [{
		$url: "?representation=s-uitest.$test&name=extend_landingCalendar&view=calendar",
		$title: "Landing Page Calendar",
	}, {
		$url: "?representation=s-uitest.$test&name=extend_xa1-color-picker",
		$title: "xa1-color-picker",
	}]

}, {
	$title: "Convergence",
	$menus: [{
		$url: "?representation=s-uitest.$test&name=tools_convergenceCleaner",
		$title: "Convergence Cleaner",
	}, {
		$title: "X3 link --> AQFA : Fields",
		$url: "/trans/x3/erp/superv/$sessions?f=AQFA"
	}, {
		$title: "X3 link --> AQFB : Array",
		$url: "/trans/x3/erp/superv/$sessions?f=AQFB"
	}, {
		$title: "X3 link --> AQFC : Picking",
		$url: "/trans/x3/erp/superv/$sessions?f=AQFC"
	}, {
		$title: "X3 link --> AQFE : windows",
		$url: "/trans/x3/erp/superv/$sessions?f=AQFE"
	}, {
		$title: "X3 link --> AQFD : Left list",
		$url: "/trans/x3/erp/superv/$sessions?f=AQFD"
	}, {
		$title: "X3 link --> AQFF : Flash",
		$url: "/trans/x3/erp/superv/$sessions?f=AQFF"
	}, {
		$title: "X3 link --> AQFG : ORG chart",
		$url: "/trans/x3/erp/superv/$sessions?f=AQFG"
	}]
}, {
	$title: "Json",
	$menus: [{
		$title: "JsonLint",
		$url: "http://jsonlint.com/"
	}, {
		$title: "Jsonformatter Curiousconcept",
		$url: "http://jsonformatter.curiousconcept.com/"
	}]
}, {
	$title: "Notification",
	$menus: [{
		$title: "test Notification",
		$url: "?representation=s-uitest.$test&name=notifications_testNotif"
	}]
}];
