// syracuse-ui/lib/field/array/gridBuilder/filter/filterHelpers.js
"use strict";
var handlers = {
	onOperatorClick: function(cell, picker, event) {
		var record = cell.record;
		var options = {
			operator: cell.operator,
			list: record.list,
			event: event,
			type: "operatorclick",
			doEvent: function() {
				if (!cell.listPopup) {
					syra_site.dialogManager.closePopups();
					syra_site.dom.empty(cell.list);
					cell._ensureFilterField();
					var operators = syra_site.filterMaker.getFieldOperators(null, cell.field, cell.$filterField);
					for (var ii = 0, jj = operators.length; ii < jj; ii++) {
						var op = operators[ii];
						var btn = syra_menus.addTextButton(syra_local["flFilter_" + op], "s-filter-popup-btn", "onFilterCellClick");
						btn.syraOnSubClick = "onFilterItemClick";
						btn.syraOperator = op;
						if (cell.operator == op) {
							btn.className += " s-selected";
						}
						btn.style.backgroundImage = "url('" + syra_site.$item.$iconPath + "page/s-filter-" + op + ".png')";
						cell.list.appendChild(btn);
					}
					cell.listPopup = syra_site.dialogManager.openPopup(record.boxParent, {
						content: record,
						slot: cell.list,
						position: {
							my: "left top",
							at: "left bottom",
							of: $(cell.layoutSlot)
						},
						onClose: function() {
							cell.listPopup = null;
						}
					});
				} else {
					cell.listPopup.close();
				}
			}
		};
		record.page.externalAdapter.onFilterEvent(options);
	},
	onFilterItemClick: function(cell, picker, event) {
		cell.setOperator(picker.syraOperator);
		cell.listPopup.close();
		cell.record.page.externalAdapter.onFilterEvent({
			operator: cell.operator,
			list: cell.record.list,
			filter: cell,
			event: event,
			type: "operatorchange",
			doEvent: function() {
				var highField = cell.record.boundFields[cell.$bind + "$High"];
				if (cell.operator === "between") {
					if (highField) {
						highField[0]._dataValue.style.display = "";
					} else {
						_loadField(cell, cell.$field, cell.$bind + "$High");
					}
				} else {
					if (highField) {
						highField[0]._dataValue.style.display = "none";
					}
					var value = cell.record.dataset[cell.$bind];
					if (cell.operator != "none") {
						if (value != null && value != "") {
							if (cell.operator == "empty" || cell.operator == "notempty") {
								var field = cell.record.boundFields[cell.$bind][0];
								field.setDataValue(null);
								cell.record._buildFilter();
							} else {
								cell.record._buildFilter();
							}
						} else {
							if (cell.operator == "empty" || cell.operator == "notempty") {
								cell.record._buildFilter();
							}
						}
					} else {
						if (value != null && value != "") {
							var field = cell.record.boundFields[cell.$bind][0];
							field.setDataValue(null);
							field.notifyFieldChange(null);
						} else {
							cell.record._buildFilter();
						}
					}
				}
			}
		});
	}


};

function _onFilterCellClick(event, btn) {
	//this is field instance
	handlers[btn.syraOnSubClick](this.filterCell, btn, event);
}

function _loadField(cell, $field, $bind) {
	cell.field = cell.record.page.loadNewItem(cell.layoutSlot.appendChild(document.createElement("div")), {
		$bind: $bind || cell.$bind,
		$isTopLabelAlignment: false,
		$isCellChild: true,
		$isFilterMode: true,
		$isMenusHiddeen: true,
		$inplace: true,
		$field: $field
	}, cell.record);
	cell._operator = syra_menus.addIconButton(syra_local.flfilter_choiceOperator, "s-list-filter-cell-picker", "onFilterCellClick");
	cell._operator.style.backgroundImage = "url('" + syra_site.$item.$iconPath + "page/s-filter-none.png')";
	cell._operator.syraOnSubClick = "onOperatorClick";
	cell.field._dataValue.insertBefore(cell._operator, cell.field._dataValue.firstChild);
	cell.field.filterCell = cell;
	cell.field.onFilterCellClick = _onFilterCellClick;
	if (!cell.list) {
		cell.list = document.createElement("div");
		cell.list.className = "s-filter-popup";
		cell.list.style.display = "none";
		cell.list.syraItem = cell.field.id;
		cell.field._dataValue.appendChild(cell.list);
	}
}

exports.renderCell = function(cell, record, layoutSlot, $field, $bind) {
	cell.$field = $field;
	cell.record = record;
	cell.$filterCode = cell.$bind = $bind;
	cell.operator = "none";
	cell.layoutSlot = layoutSlot;
	_loadField(cell, cell.$field, cell.$bind);

	// update filter cell if where params in url (issue #3022)
	if (cell.record.list.$urlParams.where) {
		var $filterBind = cell.$bind;
		var exp = syra_site.filterMaker.parseWhere(cell.record.list.$urlParams.where, cell.$bind);
		var $isRefField = false;
		// in case of reference field, appropriate $bind value is field.$reference.$value.$prop
		if ($field.$type == 'application/x-reference') {
			$filterBind = cell.field.$reference.$value.$prop;
			$isRefField = true;
		}
		if (exp && $filterBind == exp.children[0].value) {
			cell.setOperator(exp.value.text);
			var value = syra_site.filterMaker.unFormatValue(exp.value.text, exp.children[1].value);
			// update dataset and field data value
			if ($isRefField) {
				cell.record.dataset[cell.$bind] = {};
				cell.record.dataset[cell.$bind][$filterBind] = value;
			} else {
				cell.record.dataset[cell.$bind] = value;
			}
			cell.field.setDataValue(value);
		}
	}
};
