// tedious/lib/message-io.js
'use strict';

var _get = require('babel-runtime/helpers/get')['default'];

var _inherits = require('babel-runtime/helpers/inherits')['default'];

var _createClass = require('babel-runtime/helpers/create-class')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _tls = require('tls');

var _tls2 = _interopRequireDefault(_tls);

var _crypto = require('crypto');

var _crypto2 = _interopRequireDefault(_crypto);

var _events = require('events');

require('./buffertools');

var _packet = require('./packet');

var _streamParser = require('./stream-parser');

var _streamParser2 = _interopRequireDefault(_streamParser);

var ReadablePacketStream = (function (_StreamParser) {
  _inherits(ReadablePacketStream, _StreamParser);

  function ReadablePacketStream() {
    _classCallCheck(this, ReadablePacketStream);

    _get(Object.getPrototypeOf(ReadablePacketStream.prototype), 'constructor', this).apply(this, arguments);
  }

  _createClass(ReadablePacketStream, [{
    key: 'parser',
    value: _regeneratorRuntime.mark(function parser() {
      var header, _length, data;

      return _regeneratorRuntime.wrap(function parser$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            context$2$0.next = 2;
            return this.readBuffer(_packet.HEADER_LENGTH);

          case 2:
            header = context$2$0.sent;
            _length = header.readUInt16BE(2);
            context$2$0.next = 6;
            return this.readBuffer(_length - _packet.HEADER_LENGTH);

          case 6:
            data = context$2$0.sent;

            this.push(new _packet.Packet(Buffer.concat([header, data])));

          case 8:
            context$2$0.next = 0;
            break;

          case 10:
          case 'end':
            return context$2$0.stop();
        }
      }, parser, this);
    })
  }]);

  return ReadablePacketStream;
})(_streamParser2['default']);

var MessageIO = (function (_EventEmitter) {
  _inherits(MessageIO, _EventEmitter);

  function MessageIO(socket, _packetSize, debug) {
    var _this = this;

    _classCallCheck(this, MessageIO);

    _get(Object.getPrototypeOf(MessageIO.prototype), 'constructor', this).call(this);

    this.socket = socket;
    this._packetSize = _packetSize;
    this.debug = debug;
    this.sendPacket = this.sendPacket.bind(this);

    this.packetStream = new ReadablePacketStream();
    this.packetStream.on('data', function (packet) {
      _this.logPacket('Received', packet);
      _this.emit('data', packet.data());
      if (packet.isLast()) {
        _this.emit('message');
      }
    });

    this.socket.pipe(this.packetStream);
    this.packetDataSize = this._packetSize - _packet.HEADER_LENGTH;
  }

  _createClass(MessageIO, [{
    key: 'packetSize',
    value: function packetSize(_packetSize2) {
      if (arguments.length > 0) {
        this.debug.log("Packet size changed from " + this._packetSize + " to " + _packetSize2);
        this._packetSize = _packetSize2;
        this.packetDataSize = this._packetSize - _packet.HEADER_LENGTH;
      }
      return this._packetSize;
    }
  }, {
    key: 'startTls',
    value: function startTls(credentialsDetails) {
      var _this2 = this;

      var credentials = _tls2['default'].createSecureContext ? _tls2['default'].createSecureContext(credentialsDetails) : _crypto2['default'].createCredentials(credentialsDetails);

      this.securePair = _tls2['default'].createSecurePair(credentials);
      this.tlsNegotiationComplete = false;

      this.securePair.on('secure', function () {
        var cipher = _this2.securePair.cleartext.getCipher();
        _this2.debug.log("TLS negotiated (" + cipher.name + ", " + cipher.version + ")");
        _this2.emit('secure', _this2.securePair.cleartext);
        _this2.encryptAllFutureTraffic();
      });

      this.securePair.encrypted.on('data', function (data) {
        _this2.sendMessage(_packet.TYPE.PRELOGIN, data);
      });

      // On Node >= 0.12, the encrypted stream automatically starts spewing out
      // data once we attach a `data` listener. But on Node <= 0.10.x, this is not
      // the case. We need to kick the cleartext stream once to get the
      // encrypted end of the secure pair to emit the TLS handshake data.
      this.securePair.cleartext.write('');
    }
  }, {
    key: 'encryptAllFutureTraffic',
    value: function encryptAllFutureTraffic() {
      this.socket.unpipe(this.packetStream);
      this.securePair.encrypted.removeAllListeners('data');
      this.socket.pipe(this.securePair.encrypted);
      this.securePair.encrypted.pipe(this.socket);
      this.securePair.cleartext.pipe(this.packetStream);
      this.tlsNegotiationComplete = true;
    }
  }, {
    key: 'tlsHandshakeData',
    value: function tlsHandshakeData(data) {
      this.securePair.encrypted.write(data);
    }

    // TODO listen for 'drain' event when socket.write returns false.
    // TODO implement incomplete request cancelation (2.2.1.6)
  }, {
    key: 'sendMessage',
    value: function sendMessage(packetType, data, resetConnection) {
      var numberOfPackets = undefined;
      if (data) {
        numberOfPackets = Math.floor((data.length - 1) / this.packetDataSize) + 1;
      } else {
        numberOfPackets = 1;
        data = new Buffer(0);
      }

      for (var packetNumber = 0; packetNumber < numberOfPackets; packetNumber++) {
        var payloadStart = packetNumber * this.packetDataSize;

        var payloadEnd = undefined;
        if (packetNumber < numberOfPackets - 1) {
          payloadEnd = payloadStart + this.packetDataSize;
        } else {
          payloadEnd = data.length;
        }

        var packetPayload = data.slice(payloadStart, payloadEnd);

        var packet = new _packet.Packet(packetType);
        packet.last(packetNumber === numberOfPackets - 1);
        packet.resetConnection(resetConnection);
        packet.packetId(packetNumber + 1);
        packet.addData(packetPayload);
        this.sendPacket(packet);
      }
    }
  }, {
    key: 'sendPacket',
    value: function sendPacket(packet) {
      this.logPacket('Sent', packet);
      if (this.securePair && this.tlsNegotiationComplete) {
        this.securePair.cleartext.write(packet.buffer);
      } else {
        this.socket.write(packet.buffer);
      }
    }
  }, {
    key: 'logPacket',
    value: function logPacket(direction, packet) {
      this.debug.packet(direction, packet);
      return this.debug.data(packet);
    }
  }]);

  return MessageIO;
})(_events.EventEmitter);

exports['default'] = MessageIO;
module.exports = exports['default'];
