// @sage/syracuse-react/src/profile/bar/profileBarProps.ts
"use strict";
import * as Sdata from '../../dataStore/sdata';
import { Profile, IRole } from '../profile';
import { BookmarkMenu } from '../../navigation/bookmark/bookmarkMenu';
import { IEndpoint } from '../endpoint';
import { ILocalePreferences } from '../../culture/culture';
import { ITheme } from '../../style/styleSheet';

export interface IProfileBookmark {
	bookmarkDisabled: boolean,
	mainPageBookmark: BookmarkMenu,
	onBookmarked: () => void,
	onBookmarkClick: (menu: BookmarkMenu) => void,
	profile: Profile
}

export interface IProfileEndpointSelect {
	onToggleEndpointsPopup: (opened: boolean) => void,
	onSelectEndpoint: (endpoint: IEndpoint) => void
}
export interface IProfileRoleSelect {
	onToggleRolesPopup: (opened: boolean) => void,
	onSelectRole: (role: IRole) => void
}

export interface IUserTabProps extends IProfile {
	onGoToLegal: () => void,
	onGoToUser: () => void
	onUploadPhoto: () => void
}

export interface IProfileLocaleSelect {
	onSelectLocale: (local: ILocalePreferences) => void
}

export interface IHelpProps {
	onGotoHelpCenter: () => void,
	onGotoHelpShortcuts: () => void
}

export interface IProfile {
	profile: Profile
}
export interface IProfilePanelProps extends IUserTabProps, IProfileLocaleSelect {
	profile: Profile,
	onSelectTheme: (theme: ITheme) => void,
	onLogout: () => void
}

export interface ISearchProps {
	searchFieldVisible: boolean,
	onToggleSearch: () => void,
}

export interface AppConnField {
	bind: string,
	$field: Sdata.IMeta,
	title: string,
	value: any
}

export interface IAppConnProps {
	appConnFields: AppConnField[],
	onEditAppConn: () => void
}


export interface IProfileBarProps extends IProfilePanelProps, IProfileRoleSelect, IProfileEndpointSelect, IProfileBookmark, IHelpProps, ISearchProps, IAppConnProps {
	classicMode: boolean,
	onGoHome: () => void,
	onToggleSitemap: (triggerToggle: () => void) => void,
	onHeaderClick: (event: React.MouseEvent<HTMLElement>) => void
}
