// msoffice/lib/excel/excelWorksheetActions.js
"use strict"

var helpers = require('syracuse-core/lib/helpers');
var forEachKey = helpers.object.forEachKey;

exports.getPrototype = function(facetName, onSuccess, onError) {
	var result = {};
	if(facetName == "$query") {
		result.$prototype = {
            $: {
            },
            $actions: {
                refreshAll: {
                    $title: "Refresh all"
                },
                saveDocument: {
                    $title: "Save document to X3"
                }/*,
                deleteDatasource: {
                    $title: "Delete selected datasource"
                }*/
            }
        }
	    onSuccess(result);
    }
}

var _actions = {
	saveDocument: function(resource) {
        if(document.site.excelDocument.documentUrl)
    		document.site.excelDocument.publishDocument("",
            		null, null, function(data, response) {
            			// show diagnoses as server doesn't send diagnoses in this case
            			document.site.showDiagnoses({
            				$diagnoses: [{severity: "info", message: "The document has been saved"}]
            			});
            		});	
        else
            external.ShowSettingsForm();
	},
    refreshAll: function(resource) {
        document.site.excelDocument.refreshAllDatasources();
    },
    deleteDatasource: function(resource) {

    }
}

function _applyActions(resource) {
	var result = {};
	if(resource.$actions)
		forEachKey(resource.$actions, function(key, action) {
			if(!action.$isRequested)
				return;
			result = _actions[key] && _actions[key](resource);
			action.$isRequested = false;
		}); 
	return result;
}

exports.dispatcher = {
	GET: function(excelQuery, id, data, onSuccess, onError) {
    },
    PUT: function(excelQuery, id, data, onSuccess, onError) {
    	var result = null;
    	// only actions for now
    	if(data) {
    		_applyActions(data);
    	}
    	onSuccess(result);
    }
}
