// msoffice/lib/excel/excelPublisher.js
"use strict";

var helpers = require('syracuse-core/lib/helpers');
var forEachKey = helpers.object.forEachKey;

exports.getPrototype = function(reprName, facetName, onSuccess, onError){
    var result = {};
    if (facetName == "$details") {
        var docUrl = syra_site.excelDocument.documentUrl;
        // test stub >>>
        //        var docUrl = "/sdata/syracuse/collaboration/syracuse/documents('1d1bb656-da30-4bbe-95ae-343b8412d3f6')?representation=documentExcel.$details";
        //        var docUrl = "/sdata/syracuse/collaboration/syracuse/documents('67d5337a-e56e-4e75-8e7e-2e162122c798')?representation=documentExcel.$details";
        // test <<<
        if (docUrl) {
            syra_controller.loadRepresentation({
                article: null,
                segments: docUrl,
                success: function(response){
                    var repr = response.$representation;
                    repr.$prototype = repr.$prototype || {};
                    // excel specific prototype and article hack
                    repr.$article = {
                        $layout: {
                            $items: [{
                                $category: "section",
                                $layout: {
                                    $items: [{
                                        "$category": "section",
                                        "$expression": "{description} published by {owner} on {documentDate}, shared with {teams}"
                                    }]
                                }
                            }]
                        }
                    };
                    //
                    onSuccess(repr);
                }                
            });
        }
        else {
            var l = {
                $edit: {
                    $title: "Publish a generic report",
                    $type: "application/json;vnd.sage=syracuse",
                    $url: "/sdata/syracuse/collaboration/syracuse/documents/$template/$workingCopies?representation=documentExcel.$edit",
                    $method: "POST"
                }
            };
            if (syra_site.excelDocument.enableSageIntelligence) {
                l.$siReport = {
                    $title: "Publish a Sage Intelligence report",
                    $type: "application/json;vnd.sage=syracuse",
                    $url: "/sdata/syracuse/collaboration/syracuse/documents/$template/$workingCopies?representation=documentExcelSI.$edit&volumeCode=SI_REPORTS",
                    $method: "POST"
                };
                l.$siTemplate = {
                    $title: "Publish a Sage Intelligence template",
                    $type: "application/json;vnd.sage=syracuse",
                    $url: "/sdata/syracuse/collaboration/syracuse/documents/$template/$workingCopies?representation=documentExcelSI.$edit&volumeCode=SI_TEMPLATES",
                    $method: "POST"
                };
            }
            result.$prototype = {
                $properties: {},
                $links: l
            };
            onSuccess(result);
        }
    }
};

var _actions = {
    $saveDocument: function(resource){
        syra_site.excelDocument.publishDocument("", null, null, function(data, response){
            // show diagnoses as server doesn't send diagnoses in this case
            syra_site.showDiagnoses({
                $diagnoses: [{
                    severity: "info",
                    message: "The document has been saved"
                }]
            });
        });
    }
};

function _applyActions(resource){
    var result = {};
    if (resource.$actions) 
        forEachKey(resource.$actions, function(key, action){
            if (!action.$isRequested) 
                return;
            result = _actions[key] && _actions[key](resource);
            action.$isRequested = false;
        });
    return result;
}

exports.dispatcher = {
    GET: function(excelQuery, id, data, onSuccess, onError){
        var docUrl = syra_site.excelDocument.documentUrl;
        //		alert("drawbox enter:" + docUrl);
        // test stub >>>
        //        var docUrl = "/sdata/syracuse/collaboration/syracuse/documents('1d1bb656-da30-4bbe-95ae-343b8412d3f6')?representation=documentExcel.$details";
        //        var docUrl = "/sdata/syracuse/collaboration/syracuse/documents('67d5337a-e56e-4e75-8e7e-2e162122c798')?representation=documentExcel.$details";
        // test <<<
        if (docUrl) 
            syra_controller.sendRequest(null, {
                $location: {
                    $url: docUrl
                }
            }, onSuccess, onError);
    },
    PUT: function(excelQuery, id, data, onSuccess, onError){
        var result = null;
        // only actions for now
        if (data) {
            _applyActions(data);
        }
        onSuccess(result);
    }
};
