// syracuse-tablet/test/autoUI/ut.js
"use strict";

var _webdriver = require('selenium-webdriver');
var _assert = require('selenium-webdriver/testing/assert');

//var _log = true;
var _log = false;

exports.Driver = require("./lib/driver");
exports.App = require("./lib/app");
exports.SEL = require("./lib/selectors");
exports.DOM = require("./lib/dom");
exports.Config = require("./config").config;
exports.log = function(text) {
	if (_log) {
		console.log(text);
	}
};

exports.login = function() {
	return exports.App.loadApp()
		.then(function() {
			return exports.App.ensureLogin();
		});
};

exports.resolve = function() {
	var deferred = new _webdriver.promise.Deferred();
	deferred.fulfill.apply(deferred, arguments);
	return deferred.promise;
};

exports.Deferred = function() {
	return new _webdriver.promise.Deferred();
};

exports.sleep = function(time) {
	return exports.Driver.Browser().sleep(time);
};
exports.sleep$ = function(time) {
	return function() {
		return exports.Driver.Browser().sleep(time);
	};
};

exports.assert = function(subject) {
	return new _assert.Assertion(subject);
};

/** 
 * Assert that is always ok to signal this fact at the end of a process
 */
exports.assertOK = function(text) {
	var a = new _assert.Assertion(true);
	a.equals(true, text);
};

/** 
 * Assert that is always failed to signal this fact at the end of a process
 */
exports.assertFail = function(text) {
	var a = new _assert.Assertion(false);
	a.equals(true, text);
};
