// syracuse-main/html/main-tablet.js
"use strict";
var siteFactory = require('syracuse-ui/lib/tablet/tabletSite');
exports.main = function() {
	var landscape = window.orientation && (window.orientation == 90 || window.orientation == -90);
	siteFactory.load({
		$device: "tablet",
		$preferences: {
			convergenceBar: {
				$viewMode: "float",
				$isCollapsed: true
			},
			menuBar: {
				$viewMode: "dock",
				$isCollapsed: landscape ? false : true
			},
			"dash-menubar": {
				$viewMode: "dock",
				$isCollapsed: false
			}
		},
		$iconPath: "/syracuse-ui/themes/desktop/sage/images/",
		$pageTemplateUrl: "/sdata/syracuse/collaboration/syracuse/pages('{endpoint}.{representation},{target},{pageview}')?role={$role}&profile={$userprofile}",
		$userProfileUrl: "/sdata/syracuse/collaboration/syracuse/userProfiles/$template/$workingCopies?representation=userProfile.$edit&role={$role}"

	}, {
		$helpLinks: ["$documentation", "$shortcuts", "$legal"],
		$links: {
			$navigation: {
				$title: "Navigation Page",
				$url: "?representation=home.$navigation&profile={$userprofile}"
			},
			$home: {
				$title: "IPAD Welcome",
				$url: "?representation=home.$landing&profile={$userprofile}"
			},
			$help: {
				$title: "Help index",
				$url: "/help/{helpLanguage}/index.htm",
				$type: "text/html",
				$target: "help"
			},
			$support: {
				$url: "http://www.sage.com/support"
			}
		},
		$navigation: {
			$dashboard: {
				"legacyHome": {
					"$title": "Legacy Home dashboard",
					"$url": "?representation=home.$dashboard"
				},
				"devsos": {
					"$title": "Developer samples/SOS",
					"$url": "?representation=devsos.$dashboard"
				},
				"arrayTest": {
					"$title": "UI test Pages",
					"$url": "?representation=s-uitest-gridArray.$test"
				},
				"mobileHome": {
					"$title": "Mobile Home",
					"$description": "Mobile application home page",
					"$url": "/syracuse-mobile/html/mobile.html#start"
				},
				"desktophome": {
					"$title": "Desktop Home",
					"$url": "/syracuse-main/html/main.html?url=%3Frepresentation%3Dhome.%24landing"
				}
			}
		},
		$properties: {}
	});
};
if (require.main == module)
	exports.main();
