// syracuse-main/html/main-mobile.js
"use strict";
var siteFactory = require('syracuse-ui/lib/site/site');
exports.main = function() {
	siteFactory.load({
		$isMobileConfig: true,
		$iconPath: "/syracuse-ui/themes/desktop/images-general/",
		$module: "ERP X3",
		$pageTemplateUrl: "/sdata/syracuse/collaboration/syracuse/pages('{endpoint}.{representation},{target},{pageview},phone')?role={$role}",
		$userProfileUrl: "/sdata/syracuse/collaboration/syracuse/userProfiles/$template/$workingCopies?representation=userProfile.$edit&role={$role}&device=phone"

	}, {
		$links: {
			$home: {
				$title: "Welcome to Mobile Desktop Authoring",
				$url: "?representation=home.$dashboard"
			},
			$help: {
				$title: "To do: current page help",
				$url: "http://www.sage.fr/mge/logiciels-erp"
			}
		},
		$navigation: {
			$dashboard: {
				"home": {
					"$title": "Home",
					"$url": "?representation=home.$dashboard"
				},
				"samples": {
					"$title": "Developer samples/SOS",
					"$url": "?representation=samples.$dashboard"
				},
				"mobileHome": {
					"$title": "Mobile Home",
					"$url": "/syracuse-main/html/mobile.html#start"
				},
				"mobileAuthoring": {
					"$title": "Mobile Specific Authoring",
					"$url": "/syracuse-main/html/mobile-auth.html?url="
				},
				"desktopMain": {
					"$title": "Desktop main page",
					"$url": "/syracuse-main/html/main.html?url=%3Frepresentation%3Dhome.%24dashboard"
				}
			},
			$about: {
				"aboutus": {
					"$title": "About us"
				},
				"privacypolicy": {
					"$title": "Privacy Policy"
				},
				$help: {
					$title: "Help",
					$url: "http://www.sage.fr/mge/logiciels-erp/sage-erp-x3-premium-edition",
					$target: "blank"
				},
				$contactus: {
					"$title": "Contact us",
					$url: "http://www.sage.com/contactus",
					$target: "blank"
				}
			}
		},
		$properties: {}
	});
};
if (require.main == module) exports.main();
