// syracuse-main/html/main.js
"use strict";
var siteFactory = require('syracuse-ui/lib/desktop/site/desktopSite');
exports.main = function(){
    siteFactory.load({
        $layoutOptions: {
            dashboard: {
                $skin: "s-dashboard",
                section: "s-dashboard-h1",
                block: "s-dashboard-h2",
                menus: "s-h1-menus"
            },
            portlet: {
                $skin: "s-portlet",
                section: "s-portlet-h1",
                block: "s-portlet-h2",
                menus: "s-h2-menus"
            },
            page: {
                $skin: "s-page",
                section: "s-page-h1",
                block: "s-page-h2",
                menus: "s-h1-menus"
            },
            fusionpage: {
                $skin: "s-page",
                section: "s-page-h1",
                block: "s-page-h2",
                menus: "s-h1-menus"
            }
        },
        $module: "ERP X3",
        $pageTemplateUrl: "/sdata/syracuse/collaboration/syracuse/pages('{endpoint}.{representation},{target}')?role={$role}",
        $userProfileUrl: "/sdata/syracuse/collaboration/syracuse/userProfiles/$template/$workingCopies?representation=userProfile.$edit&role={$role}",
        $footerPortlets: {
            navigation: {
                $title: "Navigation",
                $url: "?endpoint=syracuse.content.demo&representation=navigation.$demo"
            },
            gadgets: {
                $title: "Gadgets",
                $url: "?endpoint=syracuse.content.demo&representation=gadgets.$demo"
            }
        }
    }, {
        $links: {
            $home: {
                $title: "Welcome",
                $url: "?representation=home.$dashboard"
            },
            $help: {
                $title: "To do: current page help",
                $url: "http://www.sage.fr/mge/logiciels-erp"
            },
            $searchLink: {
                $url: "/sdata/syracuse/search/syracuse/$search?representation=queryResult.$search&q={search}"
            }
        },
        $navigation: {
            $dashboard: {
                "home": {
                    "$title": "Home",
                    "$url": "?representation=home.$dashboard"
                },
                "mobilehome": {
                    "$title": "Mobile home",
                    "$url": "/syracuse-main/html/mobile.html?url=?representation=mobileprototype.$dashboard",
                    "$target": "blank",
                    "$tags": "MOB"
                },
                "samples": {
                    "$title": "Developper samples/SOS",
                    "$url": "?representation=samples.$dashboard"
                },
                "uioverview": {
                    "$title": "UI Overview",
                    "$url": "?representation=ui.$dashboard"
                }
            },
            $about: {
                "aboutus": {
                    "$title": "About us"
                },
                "privacypolicy": {
                    "$title": "Privacy Policy"
                },
                $help: {
                    $title: "Help",
                    $url: "http://www.sage.fr/mge/logiciels-erp/sage-erp-x3-premium-edition",
                    $target: "blank"
                },
                $contactus: {
                    "$title": "Contact us",
                    $url: "http://www.sage.com/contactus",
                    $target: "blank"
                }
            }
        },
        $: {
            search: {
                $type: "application/x-string",
                $title: "Search by key words"
            }
        }
    });
};
if (require.main == module) 
    exports.main();
