// syracuse-main/html/main-mobile.js
"use strict";
var siteFactory = require('syracuse-ui/lib/site/site');
exports.main = function(){
    siteFactory.load({
        $iconPath: "/syracuse-ui/themes/desktop/images-general/",
        $module: "ERP X3",
        $pageTemplateUrl: "/sdata/syracuse/collaboration/syracuse/pages('{endpoint}.{representation},{target},{pageview},phone')?role={$role}",
        $userProfileUrl: "/sdata/syracuse/collaboration/syracuse/userProfiles/$template/$workingCopies?representation=userProfile.$edit&role={$role}&device=phone"
    
    }, {
        $links: {
            $home: {
                $title: "Welcome",
                $url: "?representation=home.$dashboard"
            },
            $help: {
                $title: "To do: current page help",
                $url: "http://www.sage.fr/mge/logiciels-erp"
            }
        },
        $navigation: {
            $dashboard: {
                "home": {
                    "$title": "Home",
                    "$url": "?representation=home.$dashboard"
                },
                "samples": {
                    "$title": "Developer samples/SOS",
                    "$url": "?representation=samples.$dashboard"
                },
                "arrayTest": {
                    "$title": "Grid test Page",
                    "$url": "?representation=s-uitest-gridArray.$test"
                },
                "office": {
                    "$title": "Office",
                    "$url": "/msoffice/lib/excel/html/config.html?url=%3Frepresentation%3Dexcelconfig.%24dashboard"
                },
                "mobileHome": {
                    "$title": "Mobile Home",
                    "$url": "/syracuse-main/html/mobile.html#start"
                },
                "perfmon": {
                    "$title": "Performance data",
                    "$url": "/syracuse-perfmon/lib/render.html?dummy",
                    $type: "text/html",
                    $target: "blank"
                }
            },
            $about: {
                "aboutus": {
                    "$title": "About us"
                },
                "privacypolicy": {
                    "$title": "Privacy Policy"
                },
                $help: {
                    $title: "Help",
                    $url: "http://www.sage.fr/mge/logiciels-erp/sage-erp-x3-premium-edition",
                    $target: "blank"
                },
                $contactus: {
                    "$title": "Contact us",
                    $url: "http://www.sage.com/contactus",
                    $target: "blank"
                }
            }
        },
        $properties: {}
    });
};
if (require.main == module) 
    exports.main();
