// syracuse-main/html/main.js
"use strict";
var siteFactory = require('syracuse-ui/lib/site/site');
exports.main = function(){
    siteFactory.load({
        $layoutOptions: {
            page: {
                dashboard: {
                    $skin: "s-dash",
                    section: "s-dash-h1",
                    block: "s-dash-h2",
                    menus: "s-page-menus"
                },
                vignette: {
                    $skin: "s-vignette",
                    section: "s-h2",
                    block: "s-h2",
                    menus: "s-vignette-menus"
                },
                page: {
                    $skin: "s-page",
                    section: "s-h1",
                    block: "s-h2",
                    menus: "s-page-menus"
                },
                fusion: {
                    $skin: "s-page",
                    section: "s-h1",
                    block: "s-h2",
                    menus: "s-page-menus"
                }
            },
            
            searchFacet: {
                menus: "s-list-menus"
            },
            authoring: {
                $skin: "s-aw-designer-item",
                section: "s-aw-designer-h1",
                block: "s-aw-designer-h2",
                menus: "s-page-menus"
            },
            list: {
                menus: "s-list-menus",
                authoring: {
                    $skin: "s-list-designer-item",
                    section: "s-aw-designer-h1",
                    block: "s-aw-designer-h2",
                    menus: "s-page-menus"
                },
                builder: {
                    grid: {
                        section: "s-h2",
                        block: "s-h2",
                        recordMenus: "s-record-menus"
                    },
                    vgrid: {
                        section: "s-h2",
                        block: "s-h2",
                        recordMenus: "s-record-menus"
                    },
                    cards: {
                        section: "s-h2",
                        block: "s-h2",
                        recordMenus: "s-record-menus"
                    },
                    singleField: {},
                    reference: {
                        recordMenus: "s-record-menus"
                    }
                }
            
            }
        
        },
        $iconPath: "/syracuse-ui/themes/desktop/images-general/",
        $module: "ERP X3",
        $pageTemplateUrl: "/sdata/syracuse/collaboration/syracuse/pages('{endpoint}.{representation},{target},{variant}')?role={$role}",
        $userProfileUrl: "/sdata/syracuse/collaboration/syracuse/userProfiles/$template/$workingCopies?representation=userProfile.$edit&role={$role}",
        $footerVignettes: {
            navigation: {
                $title: "Navigation",
                $url: "?endpoint=syracuse.content.demo&representation=navigation.$demo"
            },
            gadgets: {
                $title: "Gadgets",
                $url: "?endpoint=syracuse.content.demo&representation=gadgets.$demo"
            }
        }
    }, {
        $links: {
            $home: {
                $title: "Welcome",
                $url: "?representation=home.$dashboard"
            },
            $help: {
                $title: "To do: current page help",
                $url: "http://www.sage.fr/mge/logiciels-erp"
            }
        },
        $navigation: {
            $dashboard: {
                "home": {
                    "$title": "Home",
                    "$url": "?representation=home.$dashboard"
                },
                "samples": {
                    "$title": "Developper samples/SOS",
                    "$url": "?representation=samples.$dashboard"
                },
                "uioverview": {
                    "$title": "UI Overview",
                    "$url": "?representation=s-uitest-fields.$test"
                },
                "uilist": {
                    "$title": "UI List Overview",
                    "$url": "?representation=s-uitest-list.$test"
                },
                "transitionTest": {
                    "$title": "Json Page test",
                    "$url": "?representation=s-uitest-json.$test"
                }
            },
            $about: {
                "aboutus": {
                    "$title": "About us"
                },
                "privacypolicy": {
                    "$title": "Privacy Policy"
                },
                $help: {
                    $title: "Help",
                    $url: "http://www.sage.fr/mge/logiciels-erp/sage-erp-x3-premium-edition",
                    $target: "blank"
                },
                $contactus: {
                    "$title": "Contact us",
                    $url: "http://www.sage.com/contactus",
                    $target: "blank"
                }
            }
        },
        $properties: {}
    });
};
if (require.main == module) 
    exports.main();
