// syracuse-main/html/main.js
"use strict";
var siteFactory = require('syracuse-ui/lib/site/site');
exports.main = function(){
    siteFactory.load({
        $iconPath: "/syracuse-ui/themes/desktop/images-general/",
        $module: "ERP X3",
        $pageTemplateUrl: "/sdata/syracuse/collaboration/syracuse/pages('{endpoint}.{representation},{target},{variant}')?role={$role}",
        $userProfileUrl: "/sdata/syracuse/collaboration/syracuse/userProfiles/$template/$workingCopies?representation=userProfile.$edit&role={$role}",
        $footerVignettes: {
            navigation: {
                $title: "Navigation",
                $url: "?endpoint=syracuse.content.demo&representation=navigation.$demo"
            },
            gadgets: {
                $title: "Gadgets",
                $url: "?endpoint=syracuse.content.demo&representation=gadgets.$demo"
            }
        }
    }, {
        $links: {
            $home: {
                $title: "Welcome",
                $url: "?representation=home.$dashboard"
            },
            $help: {
                $title: "To do: current page help",
                $url: "http://www.sage.fr/mge/logiciels-erp"
            }
        },
        $navigation: {
            $historic: {
                "histo_banks": {
                    "$title": "List of banks"
                },
                "histo_properties": {
                    "$title": "Bank properties (Boursorama)"
                },
                "histo_newAsset": {
                    "$title": "New asset"
                },
                "histo_listAsset": {
                    "$title": "List of assets"
                },
                "histo_account": {
                    "$title": "Edit bank account (BNP)"
                }
            },
            $dashboard: {
                "home": {
                    "$title": "Home",
                    "$url": "?representation=home.$dashboard"
                },
                "samples": {
                    "$title": "Developper samples/SOS",
                    "$url": "?representation=samples.$dashboard"
                },
                "arrayTest": {
                    "$title": "Array test Page",
                    "$url": "?representation=s-uitest-array.$test"
                },
                "office": {
                    "$title": "Office",
                    "$url": "/msoffice/lib/excel/html/config.html?url=%3Frepresentation%3Dexcelconfig.%24dashboard"
                },
                "mobileHome": {
                    "$title": "Mobile Home",
                    "$url": "/syracuse-main/html/mobile.html#start"
                }
            },
            $about: {
                "aboutus": {
                    "$title": "About us"
                },
                "privacypolicy": {
                    "$title": "Privacy Policy"
                },
                $help: {
                    $title: "Help",
                    $url: "http://www.sage.fr/mge/logiciels-erp/sage-erp-x3-premium-edition",
                    $target: "blank"
                },
                $contactus: {
                    "$title": "Contact us",
                    $url: "http://www.sage.com/contactus",
                    $target: "blank"
                }
            }
        },
        $properties: {}
    });
};
if (require.main == module) 
    exports.main();
