// syracuse-main/html/main-mobile.js
"use strict";
var _site = require('syracuse-ui/lib/site/site');
var MobileJSONPage = require('syracuse-ui/lib/mobile/mobileJSONPage').MobileJSONPage;

exports.main = function() {
	var site = _site.create();
	site.$device = "mobile";
	site.onMainPageChange = function($itemPage, onLoaded) {
		var self = this;
		$itemPage.boxParent = null;
		$itemPage.layoutSlot = self.body;
		self.resetMainPage();

		$itemPage.onMainPagechange = true;
		syra_pageBuilder.load({
			$itemPage: $itemPage,
			success: function(page) {
				self.mainPage = page;
				page.urlSeg = $itemPage.urlSeg;
				page.isMainPage = true;
				self.body.style.display = "";
				self.resizeItem();
				onLoaded && onLoaded(page);
			}
		});
	};
	site.drawPage = function() {
		var self = this;
		self.addDragDrop();
		$.auth = self.mobile = {
			notify: function(event, eventData, info) {
				switch (event) {
					case "pagechange":
						!this._onChangeApplying && this.resetData(eventData, info);
						break;
					case "pageload":
						!this._onChangeApplying && this.loadData(eventData, info);
						this._onChangeApplying = false;
						break;
				}
				return eventData;
			},
			applyChange: function($article) {
				if (!this.isApplyChangeDisabled) {
					var cw = self.mobilePageFrame.contentWindow;
					if (this.name && cw.$.proto.auth.applyAuth) {
						this._onChangeApplying = true;
						cw.$.proto.auth.applyAuth(this.name, $article);
					}
				}
			},
			resetData: function(eventData, info) {
				this.name = "";
			},
			loadData: function(data, info) {
				this.name = info.name;
				syra_site.onMainPageChange({
					urlSeg: syra_url.parse(info.url.origUri),
					initData: info.dataset,
					$representation: {
						$authorUrl: this.$authorUrl = info.$authorUrl,
						$prototype: info.$prototype,
						$article: data
					}
				});
			},
			toggleJsonEditor: function(show) {
				var mainPage = syra_site.mainPage;
				if (show) {
					mainPage.domItem.style.display = "none";
					this._jsonPage = new MobileJSONPage();
					syra_pageBuilder.initialize(this._jsonPage);
					this._jsonPage.layoutSlot = document.createElement("div");
					syra_site.body.appendChild(this._jsonPage.layoutSlot);
					this._jsonPage.load();
					this._jsonPage.fillEditors(mainPage.$prototype, mainPage.$item);
					this._jsonPage.isPageLoaded = true;
					syra_layout.ensureArticleVisibility(this._jsonPage);
				} else {
					mainPage.domItem.style.display = "";
					if (this._jsonPage) {
						if (this._jsonPage.layoutSlot) {
							syra_dom.remove(this._jsonPage.layoutSlot);
						}
						this._jsonPage.dispose();
						this._jsonPage = null;
					}
				}
			}

		};
		self.domItem = self.body = document.createElement("div");
		self._mobileWrapper = document.createElement("div");
		self._mobileWrapper.setAttribute("id", "s-mobile-site-wrapper");
		self.mobilePageSlot = document.createElement("div");
		self.mobilePageSlot.setAttribute("id", "s-mobile-page-slot");
		self.body.setAttribute("id", "s-site-body");
		self.body.setAttribute("tabindex", "1"); // add for receive keyboard event  for shortcut
		self._mobileWrapper.appendChild(self.mobilePageSlot);
		self._mobileWrapper.appendChild(self.body);
		self.layoutSlot.appendChild(self._mobileWrapper);

		self.mobilePageFrame = document.createElement("iframe");
		self.mobilePageFrame.className = "s-mobile-page-frame";
		self.mobilePageSlot.appendChild(self.mobilePageFrame);

		self.logon(function() {
			self.mobilePageFrame.setAttribute("src", "/syracuse-mobile/html/mobile-debug.html");
			self._onAfterLogon();
			self.layoutSlot.style.display = "";
			self.resizeItem();
		});
	};

	site._renderHeader = function() {
		this.header = document.createElement("header");
		this.header.id = "s-header";
		syra_dom.setZIndex(this.header);
		this.layoutSlot.insertBefore(this.header, this.layoutSlot.firstChild);

		this.headerBar = syra_dom.div("s-header-bar");

		// this.btns.authoring = _manager.button.add({
		/*
		this.btns.authoring = syra_button.add({
			parent: this,
			slot: this.headerBar,
			text: syra_local.site_btn_authoring,
			css: "s-header-link s-header-authoring",
			iconOnly: true,
			checkWorkingCopy: true,
			click: function() {
				var page;
				var site = this.parent;
				site.userProfile.toggle(false);
				if (site.mainPage && site.mainPage.isFusionPage) {
					page = syra_fusion.activatedBook.selectedSheet;
				} else {
					// var over = _manager.over.getMostOverPage();
					var over = syra_over.getMostOverPage();
					page = over && over.page;
				}
				page = page || site.mainPage;
				page && site.switchItemDesigner(page, true);
			}
		});
		*/

		site.userProfile.openerBtn = syra_button.add({
			parent: syra_site,
			slot: this.headerBar,
			iconOnly: true,
			text: syra_local.site_btn_opener,
			css: "s-header-link s-header-profile",
			click: function() {
				site.userProfile.toggle(true);
			}
		});
		//_manager.button.visibility(this.btns.authoring, false);
		//syra_button.visibility(this.btns.authoring, false);
		//site.userProfile.showIdentity();


		if (this.topPanel) {
			this.topPanel.addOpener(this.headerBar);

		}
		//this.addLogoutButton();

		syra_button.add({
			parent: this,
			slot: this.headerBar,
			text: "Logout", //syra_local.siteLogoutLink,
			iconOnly: true,
			css: "s-header-link s-header-help", //"s-site-logout", todo put back correct css
			click: function() {
				syra_site.logout();
			}
		});

		this.addItem(this.headerBar, {
			$bind: "$help",
			$category: "link",
			$noText: true,
			$skin: "s-header-link s-header-help"
		});

		this.header.appendChild(this.headerBar);
	};
	site.load();
};

if (require.main == module)
	exports.main();
