// syracuse-main/html/mobile.js
var siteFactory = require('syracuse-ui/lib/mobile/site/mobileSite');
exports.main = function(){
    siteFactory.load({
        $layoutOptions: {
            dashboard: {
                $skin: "s-dashboard",
                section: "s-dashboard-h1",
                block: "s-dashboard-h2",
                menus: "s-h1-menus"
            },
            portlet: {
                $skin: "s-portlet",
                section: "s-portlet-h1",
                block: "s-portlet-h2",
                menus : "s-h2-menus"
            },
            page: {
                $skin: "s-page",
                section: "s-page-h1",
                block: "s-page-h2",
                menus: "s-h1-menus"
            }
        },
        $module: "X3 Mobile",
        $pageTemplateUrl: "/sdata/syracuse/collaboration/syracuse/pages('{endpoint}.{representation},{target}')?role={$role}",
        $userProfileUrl: "/sdata/syracuse/collaboration/syracuse/userProfiles/$template/$workingCopies?representation=userProfile.$edit&role={$role}",
        $themes: {
            $site: {
                $home: "b",
                $mainAction: "a"
            },
            $messagebox: {
                $header: "e",
                $content: "e",
                $overlay: "e"
            },
            $dialogbox: {
                $header: "e",
                $content: "e",
                $page: "e",
                $footer: "e"
            },
            $defaultPage: {
                $headerBar: "b",
                $page: "a",
                $header: "a",
                $content: "a",
                $footerBar: "c"
            },
            $dashboard: {
                $headerBar: "b",
                $page: "a",
                $content: "a",
                $footerBar: "c",
                $navbar: "d"
            },
            $portletPage: {
                $page: "a",
                $content: "a"
            },
            $modal: {
                $page: "a",
                $headerBar: "a",
                $content: "a",
                $footerBar: "a"
            },
            $field: {
                $defaultField: {
                    $input: "a"
                },
                "application/x-collection": {
                    $pager: {
                        $button: "a",
                        $overlay: "a"
                    },
                    $item: "a"
                },
                "application/x-boolean": {
                    $input: "a"
                },
                "application/x-choice": {
                    $input: "a",
                    $overlay: "a"
                },
                $portletField: {
                    $header: "a",
                    $content: "d"
                }
            },
            $section: {
                $header: "a",
                $content: "a"
            },
            $sectionDashboard: {
                $header: "a",
                $content: "d"
            },
            $block: {
                $header: "a",
                $content: "a"
            },
            $portletMenus: {
                $items: "a"
            },
            $default: {
                $menus: {
                    $items: "a"
                },
                $button: "a",
                $item: "a"
            }
        },
        $icons: {
            menusIcon: "gear",
            mainAction: {
                $create: "plus",
                $edit: "s-icons-edit",
                $save: "check"
            },
            searchBtnIcon: "search",
            searchBtnIconpos: "notext",
            diagnoseFlag: {
                error: "s-diagnose-flag-error",
                warning: "s-diagnose-flag-warn",
                info: "s-diagnose-flag-info",
                fatal: "s-diagnose-flag-fatal"
            }
        }
    }, {
        $links: {
            $back: {
                $title: "Back"
            },
            $home: {
                $title: "Home",
                $url: "?representation=mobileprototype.$dashboard"
            }
            /*,
             $searchLink: {
             $url: "/sdata/syracuse/search/syracuse/$search/syracuse/collaboration/syracuse?representation=queryResult.$search&q={search}"
             },*/
        },
        $: {            /*search: {
             $type: "application/x-string",
             $title: "Search by key words"
             }*/
        }
    });
};
if (require.main == module) 
    exports.main();
