// syracuse-main/html/mobile.js
var siteFactory = require('syracuse-ui/lib/mobile/site/mobileSite');
exports.main = function(){
    siteFactory.load({
        $layoutOptions: {
            site: {
                theme: {
                    $home: "b",
                    $mainAction: "a"
                }
            },
            dialogbox: {
                theme: {
                    $header: "e",
                    $content: "e",
                    $page: "e",
                    $footer: "e"
                }
            },
            messagebox: {
                theme: {
                    $header: "e",
                    $content: "e",
                    $overlay: "e"
                }
            },
            page: {
                dashboard: {
                    $skin: "s-dashboard",
                    section: "s-dashboard-h1",
                    block: "s-dashboard-h2",
                    menus: "s-h1-menus",
                    theme: {
                        $headerBar: "b",
                        $page: "a",
                        $content: "a",
                        $footerBar: "c",
                        $navbar: "d",
                        $section: {
                            $header: "a",
                            $content: "d"
                        },
                        $block: {
                            $header: "a",
                            $content: "a"
                        }
                    }
                },
                portlet: {
                    $skin: "s-portlet",
                    section: "s-portlet-h1",
                    block: "s-portlet-h2",
                    menus: "s-h1-menus",
                    $navbar: "d",
                    theme: {
                        $page: "a",
                        $content: "a",
                        $section: {
                            $header: "a",
                            $content: "a"
                        },
                        $block: {
                            $header: "a",
                            $content: "a"
                        }
                    }
                },
                page: {
                    $skin: "s-page",
                    section: "mobile-box",
                    block: "mobile-box",
                    menus: "s-h1-menus",
                    $navbar: "d",
                    theme: {
                        $headerBar: "b",
                        $page: "a",
                        $header: "a",
                        $content: "a",
                        $footerBar: "c",
                        $section: {
                            $header: "a",
                            $content: "a"
                        },
                        $block: {
                            $header: "a",
                            $content: "a"
                        }
                    }
                }
            },
            searchFacet: {
                menus: "s-list-menus"
            },
            authoring: {
                $skin: "s-author-designer-item",
                section: "s-author-designer-item-h1",
                block: "s-author-designer-item-h2",
                menus: "s-h1-menus"
            },
            list: {
                menus: "s-list-menus",
                builder: {
                    cardList: {
                        section: "s-page-h2",
                        block: "s-page-h2",
                        recordMenus: "s-list-record-menus",
                        theme: {
                            $section: {
                                $header: "a",
                                $content: "a"
                            },
                            $block: {
                                $header: "a",
                                $content: "a"
                            }
                        }
                    },
                    singleField: {},
                    reference: {
                        recordMenus: "s-list-record-menus"
                    }
                }
            }
        },
		$iconPath: "/syracuse-ui/themes/desktop/images-general/",
        $module: "X3 Mobile",
        $pageTemplateUrl: "/sdata/syracuse/collaboration/syracuse/pages('{endpoint}.{representation},{target}')?role={$role}",
        $userProfileUrl: "/sdata/syracuse/collaboration/syracuse/userProfiles/$template/$workingCopies?representation=userProfile.$edit&role={$role}",
        $themes: {
            $field: {
                $defaultField: {
                    $input: "a"
                },
                "application/x-collection": {
                    $pager: {
                        $button: "a",
                        $overlay: "a"
                    },
                    $item: "a"
                },
                "application/x-boolean": {
                    $input: "a"
                },
                "application/x-choice": {
                    $input: "a",
                    $overlay: "a"
                },
                $portletField: {
                    $header: "a",
                    $content: "d"
                }
            },
            $portletMenus: {
                $items: "a"
            },
            $default: {
                $menus: {
                    $items: "a"
                },
                $button: "a",
                $item: "a"
            }
        },
        $icons: {
            menusIcon: "gear",
            mainAction: {
                $create: "plus",
                $edit: "s-icons-edit",
                $save: "check"
            },
            searchBtnIcon: "search",
            searchBtnIconpos: "notext",
            diagnoseFlag: {
                error: "s-diagnose-flag-error",
                warning: "s-diagnose-flag-warn",
                info: "s-diagnose-flag-info",
                fatal: "s-diagnose-flag-fatal"
            }
        }
    }, {
        $links: {
            $back: {
                $title: "Back"
            },
            $home: {
                $title: "Home",
                $url: "?representation=mobileprototype.$dashboard"
            }
            /*,
             $searchLink: {
             $url: "/sdata/syracuse/search/syracuse/$search/syracuse/collaboration/syracuse?representation=queryResult.$search&q={search}"
             },*/
        },
        $: {            /*search: {
             $type: "application/x-string",
             $title: "Search by key words"
             }*/
        }
    });
};
if (require.main == module) 
    exports.main();
