// syracuse-main/html/surveysite.js
"use strict";
var siteFactory = require('syracuse-ui/lib/site/surveysite');
debugger;
exports.main = function(){
    siteFactory.load({
        $layoutOptions: {
            page: {
                dashboard: {
                    $skin: "s-dash",
                    section: "s-dash-h1",
                    block: "s-dash-h2",
                    menus: "s-page-menus"
                },
                vignette: {
                    $skin: "s-vignette",
                    section: "s-h2",
                    block: "s-h2",
                    menus: "s-vignette-menus"
                },
                page: {
                    $skin: "s-page",
                    section: "s-h1",
                    block: "s-h2",
                    menus: "s-page-menus"
                },
                fusion: {
                    $skin: "s-page",
                    section: "s-h1",
                    block: "s-h2",
                    menus: "s-page-menus"
                }
            },
            
            searchFacet: {
                menus: "s-list-menus"
            },
            authoring: {
                $skin: "s-aw-designer-item",
                section: "s-aw-designer-h1",
                block: "s-aw-designer-h2",
                menus: "s-page-menus"
            },
            list: {
                menus: "s-list-menus",
                authoring: {
                    $skin: "s-list-designer-item",
                    section: "s-aw-designer-h1",
                    block: "s-aw-designer-h2",
                    menus: "s-page-menus"
                },
                builder: {
                    grid: {
                        section: "s-h2",
                        block: "s-h2",
                        recordMenus: "s-record-menus"
                    },
                    vgrid: {
                        section: "s-h2",
                        block: "s-h2",
                        recordMenus: "s-record-menus"
                    },
                    cards: {
                        section: "s-h2",
                        block: "s-h2",
                        recordMenus: "s-record-menus"
                    },
                    singleField: {},
                    reference: {
                        recordMenus: "s-record-menus"
                    }
                }
            
            }
        
        },
        $iconPath: "/syracuse-ui/themes/desktop/images-general/",
        $module: "ERP X3",
        $pageTemplateUrl: "/sdata/syracuse/collaboration/syracuse/pages('{endpoint}.{representation},{target},{variant}')?role={$role}",
        $userProfileUrl: "/sdata/syracuse/collaboration/syracuse/userProfiles/$template/$workingCopies?representation=userProfile.$edit&role={$role}",
        $footerVignettes: {
            navigation: {
                $title: "Navigation",
                $url: "?endpoint=syracuse.content.demo&representation=navigation.$demo"
            },
            gadgets: {
                $title: "Gadgets",
                $url: "?endpoint=syracuse.content.demo&representation=gadgets.$demo"
            }
        }
    }, {
        $properties: {}
    });
};
if (require.main == module) 
    exports.main();
