// syracuse-ui/lib/tests/pages/process/process.js
"use strict";
var helpers = require('syracuse-core').helpers;

var _processes = ["ACH05_002", "actions", "api_shape_dashed_borders", "apishape_rotations", "arrows", "layers", "line_caps_and_joints", "bezier", "bezier_shape_dashed_borders", "bezier_fill", "pattern", "pattern_and_rotation", "radial_gradient", "linear_gradient_and_transition", "linear_gradient_stop_colors", "linear_gradient_rotation_shapes", "text_rotation", "filling", "text_positionning", "shadow", "text_formatting", "img_rotations", "img_scaling", "rotations_and_deformations", "links", "links2", "demo_process", "demo_process2", "test"];

exports.scenarios = {};

exports.load = function(page) {
	for (var ii = 0, jj = _processes.length; ii < jj; ii++) {
		exports.scenarios[_processes[ii]] = function() {
			return [{
				run: function(page) {
					page.runUnitProcess(this);
				}
			}];
		};
	}
	page.$prototype.$properties.process = {
		$type: "application/x-process",
		$title: "Visual Process Player test"
	};
	page.leftCol.style.display = "none";
	page.rightCol.style.width = "100%";
	page.$menuOrientation = "horizontal";
	page.$isJsonFieldHidden = true;
	page.runUnitProcess = function(unit) {
		require.async("syracuse-ui/lib/tests/pages/process/data/" + unit.$title, function(err, module) {
			if (!page.flashImage) {
				page.flashImage = page.rightCol.appendChild(document.createElement("div"));
				page.loadNewItem(page.rightCol, {
					$bind: "process",
					$isTitleHidden: true,
					$isEditMode: false
				});
			}
			var $resource = helpers.object.clone(module.$resources, true);
			syra_dom.empty(page.flashImage);
			page.flashImage.style.display = "none";
			page.flashImage.style.overflow = "visible";
			page.flashImage.style.padding = "0px";
			page.applyChange($resource);
			if ($resource.flashImage) {
				syra_dom.empty(page.flashImage);
				page.flashImage.appendChild(document.createElement("img")).setAttribute("src", $resource.flashImage.$url);
				page.flashImage.style.display = "";
			}
		});
	};

};
