// syracuse-tablet/html/js/controls/page/panels/sortfilter/filterSortPanel.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var Base = require('syracuse-tablet/html/js/controls/page/panels/sidePanel').Klass;
var _Sort = require('syracuse-tablet/html/js/controls/page/panels/sortfilter/tabSort').Klass;
var _Filter = require('syracuse-tablet/html/js/controls/page/panels/sortfilter/tabFilter').Klass;
var locale = require('syracuse-tablet/html/js/helpers/locale');

var _templates = {
	main: '\
		<section class="s-m-side-panel {{side}}">\
			<ul class="nav nav-pills" role="tablist"> \
				{{#each tabs}}\
					<li style="width:{{width}};"><a href="#{{id}}" role="tab" data-toggle="tab">{{title}}</a></li> \
				{{/each}}\
			</ul> \
			<div class="tab-content">\
				<header>{{title}}</header>\
				<div class="btn-group btn-group-justified" role="group">\
				  <a type="button" class="btn btn-default" data-params="apply">{{apply}}</a>\
				  <a type="button" class="btn btn-default" data-params="close">{{close}}</a>\
				</div>\
				{{#each tabs}}\
					<div id="{{id}}" role="tabpanel" class="tab-pane {{type}}"/> \
				{{/each}}\
			</div> \
		</section>',
};


var _getHtml = function(name, ctx) {
	var key = name + "compiled";
	var tmpl = _templates[key];
	if (!tmpl) {
		tmpl = _templates[key] = Handlebars.compile(_templates[name]);
	}
	return tmpl(ctx);
};

var _tabDescr = [{
	klass: _Sort,
	options: {
		name: "sort",
		isActive: true,
		singleCriteria: true
	}
}, {
	klass: _Filter,
	options: {
		name: "filter",
		isActive: false
	}
}];

/**
 * Arrays Sort/Filter panel
 * Options
 * 		parentArray: 	parent array control
		filter: 		false to hide filter tab
		sort: 			false to hide sort tab
 */
var _Klass = utils.defineClass(
	function(controller, $type, panelId, options) {
		this._tabs = {};
		Base.call(this, controller, $type, panelId, options);
	}, Base, {
		destroy: function() {
			Base.prototype.destroy.call(this);
			this.options.parentArray = null;
			if (this._tabs) {
				$.each(this._tabs, function(key, value) {
					value.destroy();
				});
				this._tabs = null;
			}
		},
		getParentArray: function() {
			return this.options.parentArray;
		},
		/**
		 * Build panel and tabs html
		 */
		buildHtml: function(side) {
			var self = this;
			Base.prototype.buildHtml.call(this, side);
			self.$$elmt.addClass(self.typeName);
			if (self.$$panel) {
				self.$$panel.remove();
			}
			var ctx = {
				ctrlId: self.id,
				side: side,
				apply: locale.text("panels.array.apply"),
				close: locale.text("panels.array.close"),
				/* There's no titles for queries*/
				title: this.getParentArray().getTitle(),
				tabs: [],
			};
			var descriptions = [];
			/**
			 * We can disable filter or sort with self.options.filter = false or  self.options.sort
			 */
			_tabDescr.forEach(function(descr) {
				descr = $.extend({}, descr);
				descr.options.hidden = self.options[descr.options.name] === false;
				descriptions.push(descr);
			});
			var width = Math.round(1000 * 100 / _tabDescr.length) / 1000 + "%";
			var ids = {};
			descriptions.forEach(function(descr) {
				ids[descr.options.name] = utils.readableuid("array" + descr.options.name);
				ctx.tabs.push({
					id: ids[descr.options.name],
					title: locale.text("panels.array." + descr.options.name),
					type: descr.options.name,
					width: width
				});
			});
			self.$$panel = $(_getHtml("main", ctx));
			if (self.$$panel && self.$$panel.length) {
				self.$$elmt.append(self.$$panel);
				var nbHidden = 0;
				// Create tabs
				descriptions.forEach(function(descr) {
					var t = new descr.klass(self, ids[descr.options.name], descr.options);
					if (t.isHidden()) nbHidden++;
					self._tabs[descr.options.name] = t;
				});
				if (descriptions.length == 1 || descriptions.length === nbHidden + 1) {
					// Tab is alone
					$.each(self._tabs, function(key, value) {
						if (!value.isHidden()) {
							value.setAlone();
						}
					});
				} else {
					// Todo adjust with if some tabs are hidden
				}
			}
		},
		/**
		 * Notification whe panel is opened/closed
		 */
		notifyToggled: function(side, action, visible) {
			Base.prototype.notifyToggled.call(this, side, action, visible);
			$.each(this._tabs, function(id, tab) {
				tab.notifyToggled(side, action, visible);
			});
		},
		onResize: function() {
			$.each(this._tabs, function(id, tab) {
				tab.onResize();
			});
		}
	});

exports.Klass = _Klass;
