// syracuse-tablet/html/js/controls/page/panels/sortfilter/tabBase.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var jsutils = require('syracuse-tablet/html/js/helpers/jsutils');

/**
 * Base tab
 */
var _Klass = utils.defineClass(
	function(panel, id, options) {
		this.id = id;
		this.options = options || {};
		this.panel = panel;
		this.$$tab = this.panel.$$panel.find('[href="#' + id + '"]');
		this.$$tab.on('shown.bs.tab', jsutils.bindFn(this.onTabShown, this));
		this.$$content = this.panel.$$panel.find('#' + this.id);
		this.$$content.on("click", '[data-params]', jsutils.bindFn(this.onTabContentClick, this));
	}, null, {
		destroy: function() {
			this.panel = null;
			this.$$tab.unbind();
			this.$$tab.remove();
			this.$$tab = null;
			this.$$content.unbind();
			this.$$content.remove();
			this.$$content = null;
		},
		/**
		 * Hide the tab
		 */
		hide: function() {
			this.$$tab.parent().hide();
			this.$$content.hide();
		},
		setWidth: function(w) {
			this.$$tab.parent().width(w);
		},
		triggerActivate: function() {
			var self = this;
			if (self.$$tab.parent().hasClass("active")) {
				self.onTabShown();
			} else {
				window.setTimeout(function() {
					self.$$tab.tab('show');
				});
			}
		},
		/**
		 * Tab becomes visible
		 */
		onTabShown: function(event) {
			//Calls panel to set active tab
			this.panel.notifyActiveTab(this);
		},
		/**
		 * Build HTML
		 */
		buildHtml: function() {},
		/**
		 * Click on content
		 */
		onTabContentClick: function(evt) {
			evt.preventDefault();
			evt.stopPropagation();
			this.doContentAction(evt, $(evt.target), $(evt.target).attr("data-params"));
		},
		/**
		 * Called when user clicks on an action in tab content
		 * Action is identified by data-param attribute
		 */
		doContentAction: function(evt, $$target, params) {},
		/**
		 * Called when user clicks on Apply button
		 */
		doActionApply: function(evt, $$target) {},
		/**
		 * Panel is resized
		 */
		onResize: function(context) {},
		/**
		 * Panel is opened
		 */
		notifyOpenPanel: function() {},
		/**
		 * Panel is closed
		 */
		notifyClosePanel: function() {}
	});

exports.Klass = _Klass;
