// syracuse-tablet/html/js/controls/page/panels/sortfilter/tabFilterr.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var Base = require('syracuse-tablet/html/js/controls/page/panels/sortfilter/tabBase').Klass;
var fontUtils = require('syracuse-tablet/html/js/ui/fontUtils');
var locale = require('syracuse-tablet/html/js/helpers/locale');

var _templates = {
	main: '\
		<div class="s-m-scroll-wrapper">\
			<section id="filterlist">\
				<div id="filterListEmpty" style="display:none">{{emptyList}}</div>\
				<ul class="list-group"/>\
				<div class="btn-group btn-group-justified" role="group">\
					<a type="button" class="btn btn-default" data-params="addFilter">\
						<i class="fa fa-plus" data-params="addFilter"></i>\
						<span data-params="addFilter">{{addFilter}}</span>\
					</a>\
				</div>\
			</section>\
			<section id="filteredit">\
			</section>\
		</div>',
	items: '\
		{{#each items}}\
			<li class="list-group-item" data-params="{{id}}">\
				<span class="s-m-filter-info">\
					<span class="s-m-filter-field">{{fieldName}}</span>\
					<span class="s-m-filter-operator">{{operatorLabel}}</span>\
					<span class="s-m-filter-value">{{value}}</span>\
				</span>\
				<span class="s-m-filter ' + fontUtils.filterIcon("edit") + '" data-params="edit"></span>\
				<span class="s-m-filter ' + fontUtils.filterIcon("delete") + '" data-params="delete"></span>\
			</li>\
		{{/each}}',
	edit: '\
		<div data-s-idx="{{filterIdx}}">\
			{{#each ctrls}}\
				<div class="s-m-control s-m-field" id="">\
					<div class="s-m-title">{{ctrlTitle}}</div>\
					<div class="s-m-value edit">\
						{{{ctrlValue}}}\
					</div>\
				</div>\
			{{/each}}\
			<div class="btn-group btn-group-justified" role="group">\
				<a type="button" class="btn btn-default" data-params="cancelFilterEdit">\
					<i class="fa fa-times" data-params="cancelFilterEdit"></i>\
				</a>\
				<a type="button" class="btn btn-default" data-params="submitFilterEdit">\
					<i class="fa fa-check" data-params="submitFilterEdit"></i>\
				</a>\
			</div>\
		</div>',
	select: '\
		<select class="form-control ctrl-evt-change s-m-meta {{type}}" data-control-id="{{panelId}}">\
			{{#each options}}\
				<option value="{{value}}" {{#if selected}}selected="true"{{/if}}>{{optionLabel}}</option>\
			{{/each}}\
		</select>',
	input: '\
		<input class="s-m-meta form-control" type="text" value="{{#if filterValue}}{{filterValue}}{{/if}}">',
	editt: '<b style="color:white">Under construction</b>',
};


var _getHtml = function(name, ctx) {
	var key = name + "compiled";
	var tmpl = _templates[key];
	if (!tmpl) {
		tmpl = _templates[key] = Handlebars.compile(_templates[name]);
	}
	return tmpl(ctx);
};

/**
 * Filter tab
 */
var _Klass = utils.defineClass(
	function(panel, id, options) {
		Base.call(this, panel, id, options);
		this.$$content.append($(_getHtml("main", {
			addFilter: locale.text("panels.array.addFilter"),
			emptyList: locale.text("panels.array.emptyList")
		})));
		this.$$listSection = this.$$content.find("#filterlist");
		this.$$list = this.$$listSection.find(".list-group");
		this.$$editSection = this.$$content.find("#filteredit");
		this.$$emptyList = this.$$content.find("#filterListEmpty");
	}, Base, {
		/**
		 * Tab becomes Active - we create content
		 */
		onTabShown: function() {
			Base.prototype.onTabShown.call(this);
			this.updateFilterList();
		},
		/**
		 * Update filter list . If no filter set, display no filter information
		 **/
		updateFilterList: function() {
			this.$$emptyList.toggle(false);

			// clean current list
			this.$$list.empty();

			// We clone to keep a different copy from array one
			this.filterInfo = this.filterInfo || $.extend(true, [], this.panel.getParentArray().filterInfoGet());

			if (this.filterInfo && this.filterInfo.length > 0) {
				// clean
				this.$$list.empty();

				// append list
				var $$items = $(_getHtml("items", {
					items: this.filterInfo
				}));
				this.$$list.append($$items);
			} else {
				// no filter set
				this.$$emptyList.toggle(true);
			}
		},
		/**
		 * We remove the content on close to cancel modifications
		 */
		notifyClosePanel: function() {
			Base.prototype.notifyClosePanel.call(this);
			if (this.$$items) {
				this.$$items.remove();
				this.$$items = null;
			}
			this.filterInfo = null;
		},
		/**
		 * Save sort info
		 */
		doActionApply: function(evt, $$target, dataParams) {
			this.panel.getParentArray().filterInfoSet($.extend(true, [], this.filterInfo), true);
			return true;
		},
		/**
		 * Click on an element with data-param - edit delete...
		 */
		doContentAction: function(evt, $$target, dataParams) {
			var self = this;
			switch (dataParams) {
				case "addFilter":
					self._openFilterEdit();
					break;
				case "cancelFilterEdit":
					self._onCancelEdit();
					break;
				case "submitFilterEdit":
					self._onFilterEditSubmit();
					break;
				case "delete":
					var idx = $$target.closest("li").index();
					self._onFilterDelete(idx);
					break;
				case "edit":
					var idx = $$target.closest("li").index();
					self._openFilterEdit(idx);
					break;
				default:
					alert(dataParams);
			}
		},
		/*
		 *	When user deletes one filter in filter list
		 */
		_onFilterDelete: function(idx) {
			var self = this;
			if (self._isIndexValid(idx)) {
				self._removeFilter(idx);
				self.updateFilterList();
			}
		},
		/*
		* Adds new filter info
		* {
			fieldName: "field name",
			operator: "selected operator label",
			value: "filtering value"
		}
		*/
		_addFilter: function(newFilterInfo) {
			var self = this;
			self.filterInfo.push(newFilterInfo);
		},
		/*
		 * Remove filter
		 */
		_removeFilter: function(idx) {
			var self = this;
			if (self._isIndexValid(idx)) {
				self.filterInfo.splice(idx, 1);
			}
		},
		/*
		 * Update filter
		 */
		_updateFilter: function(idx, updateFilterInfo) {
			var self = this;
			if (self._isIndexValid(idx)) {
				self.filterInfo.splice(idx, 1, updateFilterInfo);
			}
		},
		/*
		 * If user cancels filter edit, hide form and show filters list
		 */
		_onCancelEdit: function() {
			var self = this;
			self.$$editSection.empty();
			self.$$listSection.toggle(true);
		},
		/*Build filter set/add form */
		_openFilterEdit: function(filterIdx) {
			var self = this;

			// get filter edit tab
			self.filterEditTab = self.filterEditTab || self.panel.getTabs().form;

			// set init data
			if (self._isIndexValid(filterIdx)) {
				self.filterEditTab.setInitData(self.filterInfo[filterIdx], filterIdx);
			}

			// show filter edit tab content
			self.filterEditTab.triggerActivate();
		},
		_isIndexValid: function(idx) {
			return !isNaN(parseInt(idx, 10)) && idx >= 0;
		},
		onFilterUpdate: function(filterInfo, filterIdx) {
			var self = this;

			if (self._isIndexValid(filterIdx)) {
				self._updateFilter(filterIdx, filterInfo);
			} else {
				self._addFilter(filterInfo);
			}

			self.triggerActivate();
		}
	});

exports.Klass = _Klass;
