// syracuse-tablet/html/js/controls/page/panels/sortfilter/tabFilter.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var Base = require('syracuse-tablet/html/js/controls/page/panels/sortfilter/tabBase').Klass;
var fontUtils = require('syracuse-tablet/html/js/ui/fontUtils');
var locale = require('syracuse-tablet/html/js/helpers/locale');

var _templates = {
	main: '\
		<div class="s-m-scroll-wrapper">\
			<section id="filterlist">\
				<div id="filterListEmpty" style="display:none">{{emptyList}}</div>\
				<ul class="list-group"/>\
				<div class="btn-group btn-group-justified" role="group">\
					<a type="button" class="btn btn-default" data-params="addFilter">\
						<i class="fa fa-plus" data-params="addFilter"></i>\
						<span data-params="addFilter">{{addFilter}}</span>\
					</a>\
				</div>\
			</section>\
			<section id="filteredit">\
			</section>\
		</div>',
	items: '\
		{{#each items}}\
			<li class="list-group-item" data-params="{{id}}">\
				<span class="s-m-filter-info">\
					<span class="s-m-filter-field">{{fieldName}}</span>\
					<span class="s-m-filter-operator">{{operatorLabel}}</span>\
					<span class="s-m-filter-value">{{value}}</span>\
				</span>\
				<span class="s-m-filter ' + fontUtils.filterIcon("edit") + '" data-params="edit"></span>\
				<span class="s-m-filter ' + fontUtils.filterIcon("delete") + '" data-params="delete"></span>\
			</li>\
		{{/each}}',
	edit: '\
		<div data-s-idx="{{filterIdx}}">\
			{{#each ctrls}}\
				<div class="s-m-control s-m-field" id="">\
					<div class="s-m-title">{{ctrlTitle}}</div>\
					<div class="s-m-value edit">\
						{{{ctrlValue}}}\
					</div>\
				</div>\
			{{/each}}\
			<div class="btn-group btn-group-justified" role="group">\
				<a type="button" class="btn btn-default" data-params="cancelFilterEdit">\
					<i class="fa fa-times" data-params="cancelFilterEdit"></i>\
				</a>\
				<a type="button" class="btn btn-default" data-params="submitFilterEdit">\
					<i class="fa fa-check" data-params="submitFilterEdit"></i>\
				</a>\
			</div>\
		</div>',
	select: '\
		<select class="form-control ctrl-evt-change s-m-meta {{type}}" data-control-id="{{panelId}}">\
			{{#each options}}\
				<option value="{{value}}" {{#if selected}}selected="true"{{/if}}>{{optionLabel}}</option>\
			{{/each}}\
		</select>',
	input: '\
		<input class="s-m-meta form-control" type="text" value="{{#if filterValue}}{{filterValue}}{{/if}}">',
	editt: '<b style="color:white">Under construction</b>',
};


var _getHtml = function(name, ctx) {
	var key = name + "compiled";
	var tmpl = _templates[key];
	if (!tmpl) {
		tmpl = _templates[key] = Handlebars.compile(_templates[name]);
	}
	return tmpl(ctx);
};

/**
 * Filter tab
 */
var _Klass = utils.defineClass(
	function(panel, id, options) {
		Base.call(this, panel, id, options);
		this.$$content.append($(_getHtml("main", {
			addFilter: locale.text("panels.array.addFilter"),
			emptyList: locale.text("panels.array.emptyList")
		})));
		this.$$listSection = this.$$content.find("#filterlist");
		this.$$list = this.$$listSection.find(".list-group");
		this.$$editSection = this.$$content.find("#filteredit");
		this.$$emptyList = this.$$content.find("#filterListEmpty");
	}, Base, {
		/**
		 * Tab becomes Active - we create content
		 */
		onTabShown: function() {
			Base.prototype.onTabShown.call(this);
			this.updateFilterList();
		},
		/**
		 * Update filter list . If no filter set, display no filter information
		 **/
		updateFilterList: function() {
			this.$$emptyList.toggle(false);

			// clean current list
			this.$$list.empty();

			// We clone to keep a different copy from array one
			this.filterInfo = this.filterInfo || $.extend(true, [], this.panel.getParentArray().filterInfoGet());

			if (this.filterInfo && this.filterInfo.length > 0) {
				// clean
				this.$$list.empty();

				// append list
				var $$items = $(_getHtml("items", {
					items: this.filterInfo
				}));
				this.$$list.append($$items);
			} else {
				// no filter set
				this.$$emptyList.toggle(true);
			}
		},
		/**
		 * We remove the content on close to cancel modifications
		 */
		notifyClosePanel: function() {
			Base.prototype.notifyClosePanel.call(this);
			if (this.$$items) {
				this.$$items.remove();
				this.$$items = null;
			}
			this.filterInfo = null;
		},
		/**
		 * Save sort info
		 */
		doActionApply: function(evt, $$target, dataParams) {
			this.panel.getParentArray().filterInfoSet($.extend(true, [], this.filterInfo), true);
			return true;
		},
		/**
		 * Click on an element with data-param - edit delete...
		 */
		doContentAction: function(evt, $$target, dataParams) {

			var self = this;
			switch (dataParams) {
				case "addFilter":
					self._openFilterEdit();
					break;
				case "cancelFilterEdit":
					self._onCancelEdit();
					break;
				case "submitFilterEdit":
					self._onFilterEditSubmit();
					break;
				case "delete":
					var idx = $$target.closest("li").index();
					self._onFilterDelete(idx);
					break;
				case "edit":
					var idx = $$target.closest("li").index();
					self._openFilterEdit(idx);
					break;
				default:
					alert(dataParams);
			}
		},
		/*
		 *	When user deletes one filter in filter list
		 */
		_onFilterDelete: function(idx) {
			var self = this;
			if (self._isIndexValid(idx)) {
				self._removeFilter(idx);
				self.updateFilterList();
			}
		},
		/*
		 * If user submits filter edit, hide form and show filters list
		 */
		_onFilterEditSubmit: function() {
			var self = this;
			var notValid = false;

			// check if field have been selected
			var $$fieldSelect = self.$$editSection.find(".form-control.ctrl-evt-change.s-m-meta.field");
			if ($$fieldSelect.val() == 0) {
				$$fieldSelect.toggleClass("error", true);
				notValid = true;
			}

			// check if value is filled
			var $$input = self.$$editSection.find("input");
			var value = $$input.val();
			if (!value || value === '') {
				$$input.toggleClass("error", true);
				notValid = true;
			}

			if (!notValid) {
				$$fieldSelect.toggleClass("error", false);
				$$input.toggleClass("error", false);

				var filterMapIdx = self.$$editSection.find(".form-control.ctrl-evt-change.s-m-meta.field").val() - 1;
				var operator = self.$$editSection.find(".form-control.ctrl-evt-change.s-m-meta.operator").val();

				var info = {
					fieldName: self.filtersMap[filterMapIdx].fieldName,
					operatorLabel: locale.text("panels.array.op." + operator),
					value: value,
					operator: self.$$editSection.find(".form-control.ctrl-evt-change.s-m-meta.operator").val(),
					id: self.filtersMap[filterMapIdx].id,
					prop: self.filtersMap[filterMapIdx].prop
				};

				// if we are editing an already existing filter, idx will be set, otherwise it will be null
				var idx = self.$$editSection.find("div[data-s-idx]").attr("data-s-idx");

				// update filter
				if (self._isIndexValid(idx)) {
					self._updateFilter(idx, info);
				}

				// add filter to filterInfo
				else {
					self._addFilter(info);
				}

				// update filter list
				self.updateFilterList();

				// remove filter edit view
				self._onCancelEdit();
			}
		},
		/*
		* Adds new filter info
		* {
			fieldName: "field name",
			operator: "selected operator label",
			value: "filtering value"
		}
		*/
		_addFilter: function(newFilterInfo) {
			var self = this;
			self.filterInfo.push(newFilterInfo);
		},
		/*
		 * Remove filter
		 */
		_removeFilter: function(idx) {
			var self = this;
			if (self._isIndexValid(idx)) {
				self.filterInfo.splice(idx, 1);
			}
		},
		/*
		 * Update filter
		 */
		_updateFilter: function(idx, updateFilterInfo) {
			var self = this;
			if (self._isIndexValid(idx)) {
				self.filterInfo.splice(idx, 1, updateFilterInfo);
			}
		},
		/*
		 * If user cancels filter edit, hide form and show filters list
		 */
		_onCancelEdit: function() {
			var self = this;
			self.$$editSection.empty();
			self.$$listSection.toggle(true);
		},
		/*Build filter set/add form */
		_openFilterEdit: function(filterIdx) {
			var self = this;

			// build filter edit view

			// clean
			this.$$editSection.empty();

			// build filters map if it's not set yet
			// /!\ DIFFERENT FROM filterInfo !!
			self.filtersMap = self.filtersMap || self.panel.getParentArray().filterMapGet();

			// build ctrls array (3 fields)
			var ctrlsHtml = self._buildFilterEditCtrls(filterIdx);

			// hide filter list view
			self.$$listSection.toggle(false);

			// show filter edit view
			this.$$editSection.append($(ctrlsHtml));
		},
		/*
		 * build filter edit ctrls which are :
		 * - field choice ctrl,
		 * - operator choice ctrl
		 * - value input ctrl
		 */
		_buildFilterEditCtrls: function(filterIdx) {
			var self = this;

			// build ctrls array (3 fields)
			var ctrls = [];

			// build fields select 
			var fieldsSelect = self._buildFieldCtrl(filterIdx);

			// build operators default select
			var opSelect = self._buildOperatorsCtrl(filterIdx);

			// get input value if any
			var inputValue = self._getInputValue(filterIdx);

			// field ctrl
			ctrls.push({
				ctrlTitle: locale.text("panels.array.filterForm.field"),
				ctrlValue: fieldsSelect
			});

			// operator ctrl
			ctrls.push({
				ctrlTitle: locale.text("panels.array.filterForm.operator"),
				ctrlValue: opSelect
			});

			// value ctrl
			ctrls.push({
				ctrlTitle: locale.text("panels.array.filterForm.value"),
				ctrlValue: _getHtml("input", {
					filterValue: inputValue
				})
			});

			return _getHtml("edit", {
				ctrls: ctrls,
				filterIdx: filterIdx
			});

		},
		onChange: function(evt) {
			var self = this;

			// if field selector changed, update  perators selection ctrl
			if ($(evt.target).hasClass("field")) {

				// nothing to do. user selected "select field" option
				if (evt.target.value == 0) {
					return;
				} else {
					// update operator select 
					var $$newCtrl = $(self._buildOperatorsCtrl(evt.target.value - 1, true));
					var $$ctrlToRemove = self.$$editSection.find(".form-control.ctrl-evt-change.s-m-meta.operator");
					var $$parent = $$ctrlToRemove.parent();
					$$parent.empty();
					$$parent.append($$newCtrl);
				}
			}

			// operator selector
			// handle specific operators such as "between", "empty", "not empty"
			else {
				//TODO
			}
		},
		/*
		 *	Build field choice/selection control for filter edit view
		 */
		_buildFieldCtrl: function(filterIdx) {
			// build fields select 
			var self = this;
			var options = [];

			// "select field" is first select option
			options.push({
				value: 0,
				optionLabel: locale.text("panels.array.filterForm.fieldDefault"),
				selected: self._isIndexValid(filterIdx) >= 0 ? false : true
			});

			// add fields to options array
			$.each(self.filtersMap, function(idx, val) {
				options.push({
					value: idx + 1,
					optionLabel: val.fieldName,
					selected: self._isIndexValid(filterIdx) ? (val.fieldName === self.filterInfo[filterIdx].fieldName ? true : false) : false
				});
			});

			// return select html
			return _getHtml("select", {
				options: options,
				panelId: self.id,
				type: "field"
			});
		},
		/*
		 *	Build operator choice/selection control for filter edit view
		 */
		_buildOperatorsCtrl: function(filterIdx, isMap) {
			var self = this;
			var operators = [];

			// if no filter map index, get default operators
			if (!self._isIndexValid(filterIdx)) {
				operators = self.panel.getParentArray().getOperators();
			}
			// if filter map index, get operators of the field type
			else {
				operators = self._getOperatorsByIdx(filterIdx, isMap);
			}

			// build selector
			var options = [];
			$.each(operators, function(idx, val) {
				options.push({
					value: val,
					selected: self._isIndexValid(filterIdx) ? (isMap ? (idx == 0 ? true : false) : val === self.filterInfo[filterIdx].operator ? true : false) : (idx == 0 ? true : false),
					optionLabel: locale.text("panels.array.op." + val)
				});
			});
			return _getHtml("select", {
				options: options,
				panelId: self.id,
				type: "operator"
			});
		},
		/*
		 *	Returns operators array of the appropriate field
		 */
		_getOperatorsByIdx: function(filterIdx, isMap) {
			var self = this;
			if (isMap) {
				return self.filtersMap[filterIdx].operators;
			} else {
				var filterFieldName = self.filterInfo[filterIdx].fieldName;
				var operators;
				$.each(self.filtersMap, function(idx, val) {
					if (val.fieldName === filterFieldName) {
						operators = val.operators;
						return;
					}
				});
				return operators;
			}
		},
		/*
		 * Get input value of filter if any
		 */
		_getInputValue: function(filterIdx) {
			var self = this;
			if (self._isIndexValid(filterIdx)) {
				var self = this;
				return self.filterInfo[filterIdx].value;
			}
			return null;
		},
		_isIndexValid: function(idx) {
			return !isNaN(parseInt(idx, 10)) && idx >= 0;
		}
	});

exports.Klass = _Klass;
