// syracuse-tablet/html/js/controls/page/panels/sortfilter/tabFilter.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var Base = require('syracuse-tablet/html/js/controls/page/panels/sortfilter/tabBase').Klass;
var fontUtils = require('syracuse-tablet/html/js/ui/fontUtils');
var locale = require('syracuse-tablet/html/js/helpers/locale');

var _templates = {
	filter: '<b style="color:white">Under construction</b>'
};


var _getHtml = function(name, ctx) {
	var key = name + "compiled";
	var tmpl = _templates[key];
	if (!tmpl) {
		tmpl = _templates[key] = Handlebars.compile(_templates[name]);
	}
	return tmpl(ctx);
};

/**
 * Filter tab
 */
var _Klass = utils.defineClass(
	function(panel, id, options) {
		// We clone to keep a copy
		this.filterInfo = panel.getParentArray().filterInfoGet(true);
		Base.call(this, panel, id, options);
	}, Base, {

		buildHtml: function() {
			this.$$content.append($(_getHtml("filter", {
				items: this.filterInfo
			})));
		},
		/**
		 * Save sort info
		 */
		apply: function(evt, $$target, dataParams) {
			// TODO
			this.panel.getParentArray().filterInfoSet(self.sortInfo, true);
			return true;
		},
		/**
		 * Click on an element wit data-param
		 */
		doAction: function(evt, $$target, dataParams) {
			// TODO
		}
	});

exports.Klass = _Klass;
