// syracuse-tablet/html/js/controls/page/panels/sortfilter/tabMemo.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var jsutils = require('syracuse-tablet/html/js/helpers/jsutils');
var Base = require('syracuse-tablet/html/js/controls/page/panels/sortfilter/tabBase').Klass;
var fontUtils = require('syracuse-tablet/html/js/ui/fontUtils');
var uiRect = require('syracuse-tablet/html/js/ui/rect');
var scroller = require('syracuse-tablet/html/js/helpers/scroller/scrollerHelpers');

var _templates = {
	main: '\
		<div id="memotabroot" class="s-m-scroll-wrapper">\
			<ul class="list-group"/>\
		</div>',
	items: '\
		{{#each items}}\
			<li class="list-group-item" data-params="{{@index}}">\
				<span class="s-m-memo-info"><a draggable="false" href="#" data-params="click">{{title}}</a>\
				<div style="display:none" class="s-m-value edit"><input class="s-m-meta form-control" type="text" value="{{title}}"></div>\
				</span>\
				<span class="s-m-memo ' + fontUtils.filterIcon("edit") + '" data-params="edit"></span>\
				<span class="s-m-memo ' + fontUtils.filterIcon("delete") + '" data-params="delete"></span>\
			</li>\
		{{/each}}',
	/** Add this line to add the remove icon
	 * <span class="s-m-order ' + fontUtils.sortIcon("none") + ' " data-params="none"></span>
	 */
};


var _getHtml = function(name, ctx) {
	var key = name + "compiled";
	var tmpl = _templates[key];
	if (!tmpl) {
		tmpl = _templates[key] = Handlebars.compile(_templates[name]);
	}
	return tmpl(ctx);
};
/**
 * Sort tab
 */
var _Klass = utils.defineClass(
	function(panel, id, options) {
		Base.call(this, panel, id, options);
		options.hidden = false;
		if (this._gestureMgr) {
			this._gestureMgr.destroy();
			this._gestureMgr = null;
		}
		this.$$content.append($(_getHtml("main", {})));
		// Create gesture mgr
		this.$$scrollWrapper = this.$$content.find(".s-m-scroll-wrapper");
		this.$$list = this.$$scrollWrapper.find("ul.list-group");
		this.$$list.css({
			height: "auto",
			position: "relative"
		});
		this._gestureMgr = scroller.newScroller(this.$$list, {
			direction: "v",
			valMax: 0,
			name: "SORTTAB",
			isPageScroller: false
		});
		this.$$list.on('blur', "input", jsutils.bindFn(this.onBlur, this));
		this.currentSelection = null;

	}, Base, {
		destroy: function() {
			Base.prototype.destroy.call(this);
			if (this._gestureMgr) {
				this._gestureMgr.destroy();
				this._gestureMgr = null;
			};
			this.$$edit = null;
			this.$$display = null;
			this.currentSelection = null;
		},
		onResize: function(context) {
			Base.prototype.onResize.call(this, context);
			if (this.$$content.is(':visible')) {
				this.updateGesture(context);
			}
		},
		updateGesture: function(context) {
			if (!this._gestureMgr) return;
			var wrapperRect = uiRect.elmtRect(this.$$scrollWrapper, "outer");
			var panelRect = uiRect.elmtRect(this.panel.$$elmt, "outer");
			var viewRect = wrapperRect.intersectRect(panelRect);
			if (viewRect && !viewRect.contains(wrapperRect)) {
				this._gestureMgr.init(viewRect);
			} else {
				this._gestureMgr.reset();
			}
		},
		/**
		 * We remove the content on close to cancel modifications
		 */
		notifyClosePanel: function() {
			Base.prototype.notifyClosePanel.call(this);
			if (this.$$items) {
				this.$$items.remove();
				this.$$items = null;
			}
		},
		/**
		 * Tab becomes Active - we create content
		 */
		onTabShown: function() {
			Base.prototype.onTabShown.call(this);
			this._render();
			this._setButtonsState();
		},
		_setButtonsState: function() {
			this.panel.setButtonsState(false, this.currentSelection != null); //currentSelection could be undefined
		},
		_render: function() {
			if (this.$$items) this.$$items.remove();
			// We clone to keep a different copy from array one		
			this.$$items = $(_getHtml("items", {
				items: this.panel.items,
			}));
			this.$$list.append(this.$$items);
			if (this.currentSelection != null) this._selectCurrent(this.currentSelection);
			this.updateGesture();
		},
		doContentAction: function(evt, $$target, dataParams) {
			if (!dataParams) return;
			switch (dataParams) {
				case "delete":
					var idx = $$target.closest("li").index();
					this._onMemoDelete(idx);
					break;
				case "edit":
					this._openMemoEdit($$target.closest("li"));
					break;
				case "click":
					var idx = $$target.closest("li").index();
					this._onMemoClick(idx);
					break;
				default:
					return;
			}
		},
		saveInfos: function() {
			if (this.panel.saveInfos()) {
				this.panel.savePrefs();
				this._render();
			}
		},
		_onMemoDelete: function(idx) {
			this._unSelectCurrent();
			this.panel.items.splice(idx, 1);
			this.panel.savePrefs();
			this._render();
		},
		_openMemoEdit: function(li) {
			this.$$edit = $(li).find(".s-m-value.edit");
			this.$$display = $(li).find("a");
			this._setEditMode();

		},
		_onMemoClick: function(idx) {
			if (idx != this.currentSelection) {
				this._unSelectCurrent();
			};
			this._selectCurrent(idx);
		},
		_unSelectCurrent: function() {
			this.$$list.find(".selected").toggleClass("selected");
			this.currentSelection = null;
			this._setButtonsState();
		},
		_selectCurrent: function(idx) {
			this.currentSelection = idx;
			this.$$list.find("[data-params='" + this.currentSelection + "']").toggleClass("selected");
			this._setButtonsState();
		},
		onBlur: function() {
			this._updateValue();
			this._setDisplayMode();
		},
		_setDisplayMode: function() {
			this.$$edit.toggle(false);
			this.$$display.toggle(true);
			this.$$edit = null;
			this.$$display = null;
		},
		_setEditMode: function() {
			this._unSelectCurrent();
			this.$$edit.toggle(true);
			this.$$display.toggle(false);
			this.$$edit.children().focus();
			this.$$edit.children().select();
		},
		_updateValue: function() {
			var idx = this.$$display.closest("li").index();
			var value = this.$$edit.children().val();
			this.$$display.text(value);
			this.panel.items[idx].title = value;
			this.panel.savePrefs();
		},
		doActionApply: function() {
			if (this.currentSelection !== null) {
				this.panel.getParentArray().sortInfoSet($.extend(true, [], this.panel.items[this.currentSelection].sort), false);
				this.panel.getParentArray().filterInfoSet($.extend(true, [], this.panel.items[this.currentSelection].filter), true);
				this.panel.close();
			}
		}
	});

exports.Klass = _Klass;
