// @sage/syracuse-react/src/navigation/sitemap/page/props.ts
"use strict";
import { SitemapModule, SitemapSubmodule } from '../groups';
import { SitemapBlock, SitemapMenu } from '../items';

interface IBookmarkProps {
    onMenuClick: (menu: SitemapMenu, evet: MouseEvent) => void,
    onSwitchBookmark: (menu: SitemapMenu) => void
}

interface ICollapseProps {
    onModuleCollapse: (module: SitemapModule) => void,
    onSubModuleCollapse: (submodule: SitemapSubmodule) => void,
    onMenuBlockCollapse: (menuBlock: SitemapBlock) => void
}

export interface ISitemapPageProps extends IBookmarkProps, ICollapseProps {
    userFirstName: string,
    modules: SitemapModule[],
    onModuleBodyScroll: () => void,
    onSelectModule: (module: SitemapModule) => void,
    onCollapseAllModules: (collapse?: boolean) => void
}

export interface IModuleProps extends IBookmarkProps, ICollapseProps {
    module: SitemapModule,
}

export interface ISubModuleProps extends IBookmarkProps {
    submodule: SitemapSubmodule
    onSubModuleCollapse: (submodule: SitemapSubmodule) => void,
    onMenuBlockCollapse: (menuBlock: SitemapBlock) => void
}

export interface IMenuBlockProps extends IBookmarkProps {
    collapsed?: boolean,
    menuBlock: SitemapBlock,
    onMenuBlockCollapse: (menuBlock: SitemapBlock) => void
}

export interface IMenuProps extends IBookmarkProps {
    menu: SitemapMenu
}
