// @sage/syracuse-react/src/navigation/sitemap/page/searchBar.tsx
"use strict";
import * as React from 'react';
import { Link } from '../../../link/link';
import { getLocalString } from '../../../culture/localString';

export interface ISearchBarProps {
    onSearch: (text: string) => void
}

export class SearchBar extends React.Component<ISearchBarProps, {
    value: string,
    records: any[]
}> {
    private input: HTMLInputElement;
    private value: string;
    private searchTimeout: number;

    constructor(props: any, context: any) {
        super(props, context);
        this.state = {
            value: "",
            records: []
        };
        this.value = "";
    }

    componentDidMount() {
        // this.input.value = this.value ? this.value : '';
        // this.input.focus();
    }
    componentWillUnmount() {
        //cleanup
        this.switchSearchTimer(false);
    }
    onChange = () => {
        this.value = this.input.value;
        //Run search if more than 3 character of normal ascii OR
        //Run search if  character or more not ascii (ie. chinese)
        if (this.value.length >= 3 || (this.value.length > 0 && this.value.charCodeAt(0) > 255)) {
            //Add a "debouncer" so search runs after user pauses to type
            this.switchSearchTimer(true, 200);
            // console.log("searching");
        } else {
            this.switchSearchTimer(false);
        }
        this.setState({});
    }
    onKeyPress = (ev: any) => {
        // onKeyPress = (ev: React.KeyboardEvent<any>) => {
        if (ev.key == 'Enter') {
            //Launch search
            this.props.onSearch(this.input.value);
        }
    }
    onKeyUp = (ev: React.KeyboardEvent<any>) => {
        // if (ev.key == 'Escape') {
        //     ev.preventDefault();
        //     ev.stopPropagation();
        //     this.props.onCancel();
        // }
    }
    clear = () => {
        this.input.value = "";
        this.onChange();
    }

    search() {
        if (this.value.length > 0) {
            this.props.onSearch(this.value);
        }
    }

    switchSearchTimer(on: boolean, timeout?: number) {

        //Clear / Cancel any timeout that may be running
        this.searchTimeout && clearTimeout(this.searchTimeout);
        if (on) {
            this.searchTimeout = window.setTimeout(() => {
                this.searchTimeout = null;
                this.props.onSearch(this.value);
            }, timeout);
        }
        else {
            this.searchTimeout = null;
        }
    }


    render() {
        return <div className="s_sitemap_search_bar">
            <div className="s_sitemap_search_group">
                <input className="s_sitemap_search_input"
                    placeholder="Search the sitemap"
                    type="text"
                    spellCheck={false}
                    ref={(input) => { this.input = input; }}
                    onKeyPress={this.onKeyPress}
                    onChange={this.onChange}
                    onKeyUp={this.onKeyUp} />
                <small className="s_sitemap_search_records">1 of 5</small>
                <Link className="s_sitemap_search_link nav"
                    disabled={true}
                    icon="dir_down"
                    onClick={this.search}
                    title={getLocalString("editField_cancel")} />
                <Link className="s_sitemap_search_link nav"
                    disabled={true}
                    icon="dir_up"
                    onClick={this.search}
                    title={getLocalString("editField_cancel")} />
                <Link className="s_sitemap_search_link search_close"
                    icon={this.value.length > 0 ? "close" : "magnifying_glass"}
                    onClick={this.clear}
                    disabled={this.value.length > 0 ? false : true}
                    title={getLocalString("editField_cancel")} />
            </div>
        </div>;
    }
}
