// syracuse-ui/lib/site/config.js
"use strict";
exports.$legalDataUrl = "/sdata/syracuse/collaboration/syracuse/licenseDatas?representation=licenseData.$query";
exports.$legalPrototypeUrl = "/sdata/syracuse/collaboration/syracuse/$prototypes('licenseData.$query')";
exports.$iconPath = "/syracuse-ui/themes/desktop/images/";
exports.$userProfileUrl = "/sdata/syracuse/collaboration/syracuse/userProfiles/$template/$workingCopies?representation=userProfile.$edit&role={$role}";
exports.$pageTemplateUrl = "/sdata/syracuse/collaboration/syracuse/pages('{endpoint}.{representation},{target},{pageview}')?role={$role}&profile={$userprofile}";

exports.$tecnhicalLinks = [{
	developpementMode: true,
	$bind: "changeLog",
	$title: "Change Log",
	$url: "/changelog.html",
	$type: "text/html",
	$target: "blank"
}, {
	$bind: "tablethome_v7",
	$title: "Tablet Home",
	$url: "/syracuse-main/html/main-tablet.html?url=%3Frepresentation%3Dhome.%24landing"
}, {
	$bind: "mobileHome_v7",
	"$title": "Mobile Home V7",
	"$description": "V7 Mobile application home page",
	"$url": "/syracuse-mobile/html/mobile.html#start"
}, {
	tablet: true,
	$bind: "mobileHome",
	"$title": "Mobile Home",
	"$description": "Mobile application home page",
	"$url": "/syracuse-tablet/dist/index.html"
}, {
	tablet: true,
	developpementMode: true,
	$bind: "devsos",
	$title: "Developer samples/SOS",
	$url: "?representation=devsos.$dev"
}, {
	$bind: "arrayTest",
	developpementMode: true,
	$title: "UI test Pages",
	$url: "?representation=s-uitest.$test&name=array_array&category=grid"
}, {
	developpementMode: true,
	$bind: "perfmon",
	$title: "Performance data",
	$url: "/syracuse-perfmon/public/render.html?dummy",
	$type: "text/html",
	$target: "blank"
}];



exports.size = {
	minWidth: 650,
	overResizeMinWidth: 400,
	overResizeMinHeight: 300
};
