// syracuse-ui/lib/site/dialog.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var keyHelper = require('syracuse-ui/lib/site/aside/keyHelper');

function Dialog(){
}

exports.Dialog = helpers.defineClass(Dialog, null, {
    open: function(options){
        this.id = helpers.uuid.generate();
        this._calculateMaxSize();
        if ((this.options = options).content) {
            this._content = this.options.content;
            this.$$layoutSlot = this.options.$$dialog || this.options.content.$$item;
            this.layoutSlot = this.$$layoutSlot[0];
            if (this.layoutSlot.style.position != "fixed") {
                this.layoutSlot.style.position = "absolute";
            }
            this.layoutSlot.style.display = '';
            this._ensureMode();
            if (this.options.isAutoClose !== false) {
                this._appendAutoClose(true);
            }
            if (this.options.dragSpot) {
                this._resizer = document.site.setResizable({
                    source: this,
                    slot: this.layoutSlot,
                    dragSpot: this.options.dragSpot,
                    isResizingEnabled: this.options.isResizingEnabled
                });
            }
            this._endOpen();
            if (this._resizer) {
                this._resizer.options.minHeight = this.$$layoutSlot.outerHeight(true);
                this._resizer.options.minWidth = this.$$layoutSlot.outerWidth(true);
            }
        }
        else {
            this._openDialogPage();
        }
    },
    _handleCloseOnEscape: function(){
        // handle events binding
        this._bindEvents(true);
        var close;
        if (this.options.getCloseAnchor) {
            close = this.options.getCloseAnchor();
        }
        else {
            close = document.createElement("a");
            close.className = "s-dialog-close-anchor";
            close.setAttribute('href', '#');
            this.$$layoutSlot[0].appendChild(close);
        }
        // in order for the bind to run, one element of the dom should be focused
        if (close != "noclose") {
            $(close).focus();
        }
    },
    _bindEvents: function(bind){
        var self = this;
        var namespace = ".dialog" + this.id;
        if (bind) {
            $(document).bind("keypress" + namespace, function(event){
                if (event.keyCode == keyHelper.keyCode.ESCAPE) {
                    self.close();
                    return false;
                }
            });
        }
        else {
            $(document).unbind(namespace);
        }
    },
    _appendAutoClose: function(append){
        var self = this;
        if ((self.options.$autoClose === undefined) ? true : self.options.$autoClose) {
            if (append) {
                self._autoCloseNamespace = self.$$layoutSlot.attr("id") || self.id;
                document.site.autoCloseDialogs = document.site.autoCloseDialogs || {};
                document.site.autoCloseDialogs[self.id] = self;
                self.$$layoutSlot.attr("id", self._autoCloseNamespace);
                self._autocCloseBoundary = "#" + self._autoCloseNamespace + "," + (self.options.autocCloseBoundary || ("#" + self._content.id));
                $(document).bind("click." + self._autoCloseNamespace, function(event){
                    var $$btn = $(event.target);
                    if (self.$$layoutSlot.is(":visible")) {
                        self.close();
                        /*if ($$btn.closest(self._autocCloseBoundary).length == 0) {
                         self.close();
                         }*/
                    }
                    return true;
                });
                
            }
            else {
                if (self._autoCloseNamespace) {
                    if (document.site.autoCloseDialogs) {
                        delete document.site.autoCloseDialogs[self.id];
                    }
                    $(document).unbind("click." + self._autoCloseNamespace);
                }
            }
        }
    },
    _endOpen: function(){
        this.$$layoutSlot.appendTo(document.site.layoutSlot);
        if (this._content.$dialogMode == "modal") {
            this.overlay = document.createElement("div");
            this.overlay.className = "s-overlay";
            document.site.setZIndex(this.overlay);
            document.site.body.appendChild(this.overlay);
        }
        document.site.setZIndex(this.layoutSlot);
        this.onWindowResize();
        document.site.addResizeListener(this);
        if (this.options.position) {
            this._bindScrollEvent(true);
        }
        if (this.options.onOpened) {
            this.options.onOpened(this._content);
        }
        // handle dialog close on escape key
        if (this.options.isAutoClose !== false) {
            this._handleCloseOnEscape();
        }
    },
    _bindScrollEvent: function(bind){
        var self = this;
        if (bind) {
            if (self._content.page && self._content.page.scrollview) {
                self.$$scrollview = $(self._content.page.scrollview);
                self.$$scrollview.bind("scroll." + self.id, function(evt){
                    self._moveToPosition();
                });
            }
        }
        else {
            if (self.$$scrollview) {
                self.$$scrollview.unbind("scroll." + self.id);
            }
        }
    },
    _openDialogPage: function($itemPage){
        var self = this;
        if ($itemPage || self.options.page) {
            if (self.options.page) {
                self._content = self.options.page;
                self.layoutSlot = self.options.page.layoutSlot || document.createElement("div");
                self.$$layoutSlot = $(self.layoutSlot);
                self._ensureMode();
                self._content.dialogWrapper = self;
            }
            else {
                self.layoutSlot = $itemPage.layoutSlot = document.createElement("div");
                self.$$layoutSlot = $(self.layoutSlot);
                $itemPage.onBeforeLoadPage = self.options.onBeforeLoadPage;
                $itemPage.$isResizeEnabled = false;
                $itemPage.dialogWrapper = self;
                self._content = document.site.loadNewPage($itemPage);
                self._ensureMode();
                self.layoutSlot.className = self.$skin + "-page-container";
                self._content._header.className = self.$skin + "-page-head";
                self._content._body.className = self.$skin + "-page-body";
                self._content.dataSlot.className = self.$skin + "-page-data";
                self._content._item.className = self.$skin + "-page";
                self._content._item.style.display = "none";
            }
            if (self._content.menuBar) {
                if (!$itemPage.externalAdapter || !$itemPage.externalAdapter.isDialogWithMenu || !$itemPage.externalAdapter.isDialogWithMenu($itemPage.$urlParts.$facet)) {
                    self._content.menuBar.hideBar();
                }
            }
            self.options.$autoClose = self.options.$autoClose || false;
            self._appendAutoClose(true);
            self.appendCloseButton(self._content._header);
            if (self._content.$facet == "$lookup") {
                self._content._body.style.overflow = "hidden";
            }
            else {
                self._content._body.style.overflow = "auto";
            }
            
            // self._content._body.style.height = "0px";
            if (self.options.onValidate) {
                self._appendOkButton();
            }
            if (self.options.onSelectRecord) {
                self._content.onSelectRecord = function(selectedRecords){
                    if (self.options.onSelectRecord(selectedRecords, self._content) !== false) {
                        self.close();
                    }
                };
            }
            self._content._item.style.display = "";
            
            self._endOpen();
            self._resizer = document.site.setResizable({
                source: self,
                slot: self.layoutSlot,
                dragSpot: self._content._header,
                onResize: function(resizer, moving){
                    if (moving) {
                        self._onResizeContentPage();
                    }
                }
            });
            
        }
        else {
            if (self.options.$itemPage) {
                self._openDialogPage(self.options.$itemPage);
            }
            else {
                if (self.options.$method && (self.options.$method != "GET")) 
                    document.controller._postQuery(self.options, null, self.options.article, function($location, data){
                        document.controller.loadRepresentation(self.options.article, $location.$url, function($itemPage){
                            self._openDialogPage($itemPage);
                        });
                    });
                else 
                    document.controller.loadRepresentation(self.options.article, self.options.$url, function($itemPage){
                        self._openDialogPage($itemPage);
                    });
            }
        }
    },
    _onResizeContentPage: function(){
        this._content._item.style.width = this.$$layoutSlot.width() + "px";
        //this._content.fitPageToContainer(true);
		this._content.onWindowResize(true);
		if (this.lookupList) {
            this._content._body.style.height = (this._content.$$scrollview.height() - $(this._content._header).outerHeight(true)) + "px";
            this.lookupList.onWindowResize();
        }
    },
    onFillLookupList: function(){
        if (this._content) {
            var diff = this.lookupList.builder._bodySlot.scrollHeight - this.lookupList.builder._$$bodySlot.height();
            if (diff) {
                this.layoutSlot.style.height = Math.min(this._maxHeight, this.$$layoutSlot.outerHeight(true) + diff) + "px";
                this._onResizeContentPage();
            }
        }
    },
    _ensureMode: function(){
        this._content.$dialogMode = this.options.$dialogMode || "modal";
        this.$skin = "s-" + (this._content.$dialogMode.indexOf("mod") >= 0 ? "modal" : this._content.$dialogMode);
    },
    close: function(validated, dispose){
        if (!this.disposed) {
            if (this.options && this.options.onClose ? (this.options.onClose(validated !== true, dispose) !== false) : true) {
                document.controller.disposeObject(this);
            }
        }
    },
    _appendOkButton: function(){
        var self = this;
        var btn = document.createElement("a");
        btn.className = self.$skin + "-page-ok";
        self._content._header.appendChild(btn);
        self.$$okBtn = $(btn).bind("click", function(){
            if (self.options.onValidate(self._content) !== false) {
                self.close(true);
            }
            return false;
        });
    },
    appendCloseButton: function(header){
        var self = this;
        var btn = document.createElement("a");
        btn.className = self.$skin + "-page-close";
        self.$$closeBtn = $(header.appendChild(btn)).bind("click", function(){
            self.close();
            return false;
        });
    },
    onWindowResize: function(onBoxToggle){
        if (this.overlay) {
            this.overlay.style.width = document.site.$$layoutSlot.outerWidth() + "px";
            this.overlay.style.height = document.site.$$layoutSlot.outerHeight() + "px";
            if (!this.isDeactivated) {
                if (this._content.layoutSlot.style.display !== "none") // manage of multiple window under modal
                    this.overlay.style.display = "";
            }
        }
        // in case of too fast move, dialog can be disposed too quickly
        // for example, if user opens the combo and makes a tabulation very fast, 
        // the dialog is disposed before onwindowresize is called (hack to prevent javascript error)
        if (!this.disposed) {
            this._calculateMaxSize();
            this._moveToPosition(onBoxToggle);
        }
    },
    _calculateMaxSize: function(){
        this._maxWidth = document.site.$$body.width();
        this._xMargin = this._maxWidth * 0.2;
        this._maxWidth = this._maxWidth - (this._xMargin * 2);
        this._yMargin = document.site._$$header.outerHeight(true);
        this._maxHeight = document.site.$$body.height() - (this._yMargin * 2);
        if (this._resizer) {
          //  this._resizer.maxWidth = this._maxWidth;
            this._resizer.maxHeight = this._maxHeight;
        }
        
    },
    _moveToPosition: function(onBoxToggle){
        console.log("move to position");
        if (this.options.position) {
            this.options.$$dialog.position({
                my: this.options.position.my,
                at: this.options.position.at,
                of: this.options.position.of
            });
        }
        else {
            var $$body = document.site.$$body;
            if (!this.draggedPosition) {
            
                this.options.$dialogSize = "content";
                
                if (this.options.$dialogSize == "content") {
                    if (this.$$layoutSlot.outerWidth(true) > this._maxWidth) {
                        this.layoutSlot.style.width = this._maxWidth + "px";
                    }
                    if (this.$$layoutSlot.outerHeight(true) > this._maxHeight) {
                        this.layoutSlot.style.height = this._maxHeight + "px";
                    }
                    this.layoutSlot.style.left = (($$body.width() - this.$$layoutSlot.outerWidth(true)) / 2) + "px";
                    this.layoutSlot.style.top = (($$body.height() - this.$$layoutSlot.outerHeight(true)) / 2) + "px";
                }
                else {
                    var left, top;
                    var width = $$body.width();
                    if (this.options.$maxWidth) {
                        width = Math.min(width, this.options.$maxWidth);
                    }
                    var margin = {
                        left: (width * 0.2),
                        top: $$body.height() * 0.1
                    };
                    left = margin.left;
                    top = $$body.offset().top + margin.top;
                    var width = width - (margin.left * 2);
                    if (width < 0) {
                        left = 0;
                        width = $$body.width();
                    }
                    this.layoutSlot.style.width = width + "px";
                    var height = $$body.height() - (margin.top * 2);
                    if (height < 0) {
                        top = 0;
                        height = $$body.height();
                        //  this.layoutSlot.style.height = height + "px";
                    }
                    this.layoutSlot.style.height = height + "px";
                    this.layoutSlot.style.left = left + "px";
                    this.layoutSlot.style.top = top + "px";
                }
            }
            if (this.options.$dialogSize != "content") {
                if (this._content.$$body) {
                    this._content._body.style.height = (this._content.$$scrollview.height() - $(this._content._header).outerHeight(true)) + "px";
                    /* var body = this._content.$$body[0];
                     top = 0;
                     var start = body;
                     while (start && start != this.layoutSlot) {
                     top += start.offsetTop;
                     start = start.offsetParent;
                     }
                     body.style.height = (this.$$layoutSlot.height() - top) + "px";
                     if (this._content.onWindowResize) {
                     this._content.onWindowResize();
                     }*/
                }
            }
        }
    },
    dispose: function(){
        this._bindEvents(false);
        if (this.overlay) {
            document.site.removeDomChild(this.overlay);
        }
        document.site.removeResizeListener(this);
        if (this._resizer) {
            this._resizer.dispose();
            this._resizer = null;
        }
        if (this.$$closeBtn) {
            this.$$closeBtn.unbind();
        }
        if (this.$$okBtn) {
            this.$$okBtn.unbind();
        }
        this._bindScrollEvent(false);
        if (this._content) {
            delete this._content.dialogWrapper;
        }
        if (this.options) {
            this.options.dragSpot = this.options.page = this.options.onBeforeLoadPage = null;
            if (this.options.boxParent) {
                this.options.boxParent.closeDialog(this);
            }
            this._appendAutoClose(false);
            if (this.options.content) {
                //feree external content
                delete this.options.content;
                if (this.$$layoutSlot) {
                    this.$$layoutSlot.detach();
                }
            }
            else {
                //content is internal
                document.controller.disposeObject(this._content);
                if (this.$$layoutSlot) {
                    this.$$layoutSlot.empty().remove();
                }
            }
            this.options.article = this.options.onValidate = this.options.onSelectRecord = null;
        }
        this.overlay = this.$$scrollview = this._content = this.lookupList = null;
    }
});
