// syracuse-ui/lib/site/disconnectUtils.js
"use strict";
var locale = require('syracuse-core/lib/locale');

exports.onBeforeUnload = function(event){
    var site = document.site;
    
    var localize = locale.resources(module)();
    // if some convergence sessions opened, user is warned
    if (document.controller && document.controller.fusionGateway && document.controller.fusionGateway._books) {
        site.books = document.controller.fusionGateway._books;
        var booksNb = site.books.length;
        if (booksNb > 0) {
            // alert message is shown according to the browser
            var alertMsg = localize.d_unload_alert.replace("{booksNb}", booksNb);
            switch (getBrowserName().toLowerCase()) {
                case "firefox": // display alert text in diagnoses panel before browser alert box
                    site.showDiagnoses({
                        $diagnoses: [{
                            "severity": "error",
                            "message": alertMsg
                        }],
                        autoHide: {
                            "autoHide.autoHideTimeOut": 1000
                        }
                    });
                    return alertMsg;
                case "chrome": // display alert text in browser alert box
                case "safari":
                case "msie":
                    var ee = event || window.event;
                    if (ee) {
                        ee.returnValue = alertMsg;
                    }
                    return alertMsg;
                default: // both alert display methods 
                    site.showDiagnoses({
                        $diagnoses: [{
                            "severity": "error",
                            "message": alertMsg
                        }]
                    });
                    var ee = event || window.event;
                    if (ee) {
                        ee.returnValue = alertMsg;
                    }
                    return alertMsg;
            }
        }
    }
}



exports.onUnload = function(){
    //var site = document.site;
    //if (site.books && site.books.length > 0) {
    // send ACK_TERM to all convergence sessions
    //}
}



exports.onLogoutUnload = function(){
    //TODO
    // send ACK_TERM. If OK --> keep login out. If NOK, stay in the page. 
}

function getBrowserName(){
    var N = navigator.appName, ua = navigator.userAgent;
    var M = ua.match(/(opera|chrome|safari|firefox|msie)\/?\s*(\.?\d+(\.\d+)*)/i);
    M = M ? M[1] : N;
    return M;
}
