// syracuse-ui/lib/site/site.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var _tip = require('syracuse-ui/lib/dialog/tip');
var _dialogManager = require('syracuse-ui/lib/dialog/dialogManager');
var ExternalAdapter = require("./aside/externalAdapter").ExternalAdapter;

var _diagnoseCenter = require('syracuse-ui/lib/diagnoses/diagnoseCenter');
var _dragDropManager = require("syracuse-ui/lib/site/tools/dragDropManager");
var RawPage = require("syracuse-ui/lib/page/rawPage").RawPage;
var _controller = require('syracuse-ui/lib/controller/controller');
var _workingCopy = require('syracuse-ui/lib/controller/workingCopy');
var eventListener = require('syracuse-ui/lib/events/eventListener');
var locale = require('syracuse-core/lib/locale');

var pageLoader = require("syracuse-ui/lib/page/pageLoader");
var pageHelper = require('syracuse-ui/lib/page/pageHelper');

var _trackersManager = require('syracuse-ui/lib/tracker/trackersManager');
var _deltaManager = require('syracuse-ui/lib/site/tools/deltaManager');
var FusionGateway = require('syracuse-ui/lib/fusion/fusionGateway').FusionGateway;
var UserProfile = require('./userProfile').UserProfile;
var _layoutValidator = require("syracuse-ui/lib/article/layout/layoutValidator");
var _layoutUpdater = require("syracuse-ui/lib/article/layout/layoutUpdater");
var expressionMaker = require('./tools/expressionMaker');
var localizer = require('syracuse-ui/lib/localization/localizer');
var _url = require('syracuse-ui/lib/site/tools/url');
var filterMaker = require('syracuse-ui/lib/site/filter/filterMaker');
var _domHelper = require('syracuse-ui/lib/site/tools/domHelper');
var uiLocker = require('syracuse-ui/lib/site/tools/uiLocker');
var _maximizer = require('syracuse-ui/lib/site/tools/maximizer');
var _bookmarksManager = require("syracuse-ui/lib/page/bookmarks/bookmarksManager");
var _store = require('syracuse-ui/lib/site/tools/store');
var _helpCenter = require('syracuse-ui/lib/site/help/helpCenter');
var _menusHelper = require('syracuse-ui/lib/menus/menusHelper');
var _fieldHelper = require('syracuse-ui/lib/field/helpers/fieldHelper');
var _topPanel = require('syracuse-ui/lib/site/aside/topPanel');
var _searcher = require('syracuse-ui/lib/search/searcher');
var _qualityController = require('syracuse-ui/lib/site/quality/qualityController');
var jsurl = require('jsurl/lib/jsurl');
var _widgetsLibrary = require('./widgetsLibrary');

var _styleSheetPath = "/syracuse-ui/themes/desktop/";
var _styleSheets = {};
var _themeStyleSheet;

function _unloadStyleSheet(name) {
	var link = _styleSheets[name];
	link && syra_site.documentHeader.removeChild(link);
	delete _styleSheets[name];
}

function _ensureSectionVisibility(section) {
	if (section.isVignetteField) {
		return true;
	}
	var isVisible = false;
	if (section.$item && section.$item.$isEmptyVisible) {
		isVisible = true;
	} else {
		if (section.childrenSection) {
			for (var ii = 0, jj = section.childrenSection.length; ii < jj; ii++) {
				if (_ensureSectionVisibility(section.childrenSection[ii])) {
					isVisible = true;
				}
			}
		}
		if (!isVisible && section.boxChildItems.length) {
			for (var ii = 0, jj = section.boxChildItems.length; !isVisible && ii < jj; ii++) {
				isVisible = !section.boxChildItems[ii].isInvisible;
			}
		}
	}
	if (isVisible == section.isInvisible || section.isInvisible === undefined) {
		section.isInvisible = !isVisible;
		if ((section.isInvisible !== section.$isHidden) || (!section.isInvisible && !section.$isHidden && section.isItemHidden)) {
			section.showItem(!(section.isInvisible || section.$isHidden));
		}
	}
	return isVisible;
}



function Site() {}

exports.Site = helpers.defineClass(Site, RawPage, {
	ensureArticleVisibility: function(article, resizeIfNeed, boxParent) {
		if (article.page == article) {
			if (!article.isPageLoaded || article._isDataChanging) {
				return;
			}
			resizeIfNeed = true;
		}
		if (article.childrenSection) {
			for (var mm = 0, kk = article.childrenSection.length; mm < kk; mm++) {
				if (!boxParent || boxParent.isChild(article.childrenSection[mm])) {
					_ensureSectionVisibility(article.childrenSection[mm]);
				}
			}
		}
		if (article.records) {
			for (var ii = 0, jj = article.records.length; ii < jj; ii++) {
				this.ensureArticleVisibility(article.records[ii]);
			}
		}
		var layoutContent = (boxParent || article).layoutContent;
		layoutContent && this.layoutValidator.validate(layoutContent, true);
		resizeIfNeed && article.resizeArticle && article.resizeArticle(true, boxParent);
	},
	loadStyleSheet: function(name) {
		if (!_styleSheets[name]) {
			var link = _styleSheets[name] = document.createElement("link");
			link.syraName = name;
			link.setAttribute("rel", "stylesheet");
			link.setAttribute("type", "text/css");
			link.setAttribute("href", _styleSheetPath + name);
			if (_themeStyleSheet) {
				this.documentHeader.insertBefore(link, _themeStyleSheet);
			} else {
				this.documentHeader.appendChild(link);
			}
			return link;
		}
	},
	applyTheme: function(theme) {
		if (theme && theme.cssFiles) {
			var name = theme.cssFiles[0].path.toLowerCase();
			if (name.indexOf(".css") < 0) {
				name = "custom/" + name + ".css";
				if (_themeStyleSheet) {
					if (_themeStyleSheet.syraName == name) {
						return;
					}
					_unloadStyleSheet(_themeStyleSheet.syraName);
				}
				_themeStyleSheet = null;
				_themeStyleSheet = this.loadStyleSheet(name);
			}
		} else {
			_themeStyleSheet && _unloadStyleSheet(_themeStyleSheet.syraName);
			_themeStyleSheet = null;
		}
	},
	getBrowserIdentity: function() {
		if (!this._browserIdentity) {
			var M = navigator.userAgent.match(/(opera|chrome|safari|firefox|msie)\/?\s*(\.?\d+(\.\d+)*)/i);
			this._browserIdentity = {
				name: (M ? M[1] : navigator.appName).toLowerCase()
			};
			var tem;
			if (M && (tem = navigator.userAgent.match(/version\/([\.\d]+)/i)) != null) {
				M[2] = tem[1];
			}
			this._browserIdentity.info = M ? [M[1], M[2]] : [navigator.appName, navigator.appVersion, '-?'];
			this._browserIdentity.version = this._browserIdentity.info[1];
			this._browserIdentity.isMSIE = this._browserIdentity.name == "msie";
		}
		return this._browserIdentity;
	},
	checkMainPageBookmarkStatus: function() {
		this.bookmarksManager && this.bookmarksManager.checkMainPageStatus();
	},
	setSpecificAttributes: function(dom, specificity) {
		if (dom) {
			switch (specificity) {
				case "disableSpellCheck":
					dom.setAttribute("spellcheck", false);
					break;
				case "disableAutoCorrect":
					dom.setAttribute("autocorrect", "off");
					break;
				default:
					dom.setAttribute("spellcheck", false);
					dom.setAttribute("autocorrect", "off");
			}
		}
	},
	randomStr: function(m) {
		var m = m || 9,
			s = '',
			r = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
		for (var i = 0; i < m; i++) {
			s += r.charAt(Math.floor(Math.random() * r.length));
		}
		return s;
	},
	logout: function(doPost) {
		if (!doPost && this.fusionGateway) {
			this.fusionGateway.onUserActionUnload(function() {
				syra_site.logout(true);
			}, null, "logout");
			return;
		}
		_controller.postQuery({
			$title: syra_local.siteLogout,
			$method: "POST",
			$url: "/logout",
			httpHeaders: {
				authorization: "basic"
			}
		}, null, null, function($location, data) {
			syra_site.hasDoneLogOut = true;
			window.open($location.$url, "_self");
		}, function($diagnoses) {
			syra_diagnose.showDiagnoses({
				$diagnoses: $diagnoses
			});
			if (!syra_site.mainPage || syra_site.mainPage.disposed) {
				syra_menus.click.home();
			}
		});
	},
	onMenuClick: function(menuItem) {
		switch (menuItem.$bind) {
			case "$designerDocumentation":
			case "$documentation":
				if (!menuItem.$documentationUrl) {
					menuItem.$documentationUrl = menuItem.$sourceUrl;
				}
				var edp = this.userProfile.getSelectedEndpoint();
				menuItem.$sourceUrl = menuItem.$documentationUrl + (edp ? ("&ep=" + edp.$uuid) : "");
				break;
			case "$help":
				_helpCenter.openPopup(menuItem);
				return false;
			case "$legal":
				_helpCenter.openLegalPage(menuItem);
				return false;
			default:
				if (this.bookmarksManager && menuItem.$bind && menuItem.$bind.split("-")[0] === "bookmark") {
					this.bookmarksManager.onLinkClick(menuItem);
					return false;
				}
				break;
		}
		return true;
	},
	dispose: function() {
		eventListener.dispose();
		this.fusionGateway && this.fusionGateway.dispose();
		this.bookmarksManager && this.bookmarksManager.dispose();
		_trackersManager.dispose();
		_topPanel.dispose();
		_dialogManager.dispose();
		this.uiLocker && this.uiLocker.dispose();
		_dragDropManager.dispose();
		_diagnoseCenter.box.dispose();
		_helpCenter.dispose();
		RawPage.prototype.dispose.call(this);
		syra_store && syra_store.dispose();
		var keys = Object.keys(window);
		for (var ii = 0, jj = keys.length; ii < jj; ii++) {
			var key = keys[ii];
			if (key.indexOf("syra_") >= 0) {
				delete window[key];
			}
		}
		_themeStyleSheet = _styleSheets = null;
	},
	_focusBox: function(box, hasFocus) {
		if (box) {
			var target = box.layoutSlot || box.body;
			if (!target && box.domItem) {
				target = box.domItem;
			}
			if (target) {
				if (box != box.page) {
					this.dom.toggleClass(target, "s-selected-box", hasFocus);
				}
			}
		}
	},

	logon: function(onAfterLogon) {
		var self = this;
		var prevdataset;
		if (self.userProfile) {
			prevdataset = self.userProfile.buildWCContent();
			self.userProfile.dispose(self.userProfile);
			self.userProfile = null;
		}
		self.userProfile = new UserProfile();
		self.userProfile.dataset = {}; //important
		self.userProfile.isSiteRegisterDisabled = true;
		self.pageLoader.initialize(self.userProfile);
		self.userProfile.onLogonProcessing = true;


		var urlSeg = _url.history.getUrlSegment();
		var $userProfileUrl = self.userProfile.getLogonUrl();
		var $userProfileSeg = _url.history.getUrlSegment($userProfileUrl);
		var profile = urlSeg.params && urlSeg.params.profile;
		var selectedEndpoint, selectedRole;
		if (profile) {
			profile = jsurl.parse(profile);
			selectedEndpoint = "\'" + profile.ep + "\'";
			selectedRole = "\'" + profile.role + "\'";
		}
		if (!selectedEndpoint) {
			if (!urlSeg.hasNoEndpoint) {
				if (urlSeg.endpointSeg.application != "syracuse" && ["collaboration", "search"].indexOf(urlSeg.endpointSeg.contract) < 0) {
					if (urlSeg.endpointSeg.dataset != $userProfileSeg.dataset) {
						selectedEndpoint = "dataset eq \'" + urlSeg.endpointSeg.dataset + "\'";
					}
				}
			}
		}
		if (!prevdataset && (selectedEndpoint || selectedRole)) {
			var $baseUrl = _url.getBaseUrl($userProfileSeg);
			prevdataset = {
				$etag: 0
			};
			if (selectedRole) {
				prevdataset.selectedRole = {
					$url: $baseUrl + "/roles(" + selectedRole + ")"
				};
			}
			if (selectedEndpoint) {
				prevdataset.selectedEndpoint = {
					$url: $baseUrl + "/endPoints(" + selectedEndpoint + ")"
				};
			}
		}
		syra_workingCopy.load({
			menu: {
				$url: $userProfileUrl,
				sendData: prevdataset
			},
			article: self.userProfile,
			callback: function() {
				if (!self.hasLogonFailed) {
					self.userProfile.onLogonSucees();
					locale.setCurrent(function(err) {
						window.syra_local = locale.resources(module)();
						self.dataset.language = self.dataset.helpLanguage = self.localizer.getLanguage() || "";
						var ids = ["designerDocumentation", "documentation", "shortcuts", "legal", "help", "home", "navigation"];
						for (var ii = ids.length - 1; ii >= 0; ii--) {
							var $menu = self.$menus["$" + ids[ii]];
							if ($menu) {
								$menu.$title = window.syra_local["site_link_" + ids[ii]] || "";
							}
						}
						self._renderHeader();
						_topPanel.showUserIdentity(self.userProfile);
						syra_quality.onAfterLogon(onAfterLogon);
						delete self.userProfile.onLogonProcessing;
					}, self.userProfile.currentLangCode, self.userProfile.dataset.selectedLocale);
				}
			}
		});

	},
	getSessIdCookieName: function() {
		return "syracuse.sid." + window.location.port;
	},
	getSessionId: function() {
		return helpers.http.parseCookie(document.cookie)["syracuse.sid." + window.location.port] || "";
	},
	unload: function() {
		this.dispose();
	},
	disableRequestSurvey: function(disabled) {
		this._isdRequestSurveyDisabled = disabled;
		if (this.requestSurvey) {
			if (disabled) {
				this.dom.toggleClass(this.requestSurvey, "s-on", false);
			} else {
				this.notifyRequestSurvey(this.requestSurvey.syraCount);
			}
		}
	},
	notifyRequestSurvey: function(count) {
		if (!this.requestSurvey) {
			this.requestSurvey = document.createElement("div");
			this.requestSurvey.id = "s-site-request";
			this.layoutSlot.appendChild(this.requestSurvey);
		}
		count = Math.max(count || 0, 0);
		syra_quality.logRequestCount(count);
		this.requestSurvey.textContent = count;
		if (!this._isdRequestSurveyDisabled) {
			if (this.requestSurvey.syraCount !== count) {
				if (count && !this.requestSurvey.syraCount) {
					this.dom.toggleClass(this.requestSurvey, "s-on", true);
				}
				if (!count) {
					this.dom.toggleClass(this.requestSurvey, "s-on", false);
				} else {
					this.dom.setZIndex(this.requestSurvey, true);
				}
			}
		}
		this.requestSurvey.syraCount = count;
	},
	toggleMainPage: function(page, show) {
		_dialogManager.activateAll(page, show);
		if (show && page.domItem && page.displayed !== true) {
			page.displayed = show;
			this.refreshBrowserTitle(this.mainPage = page);
			(page.layoutSlot = this.body).appendChild(page.domItem);
			page.resizeArticle();
		} else {
			if (!show && page.displayed !== false) {
				page.displayed = show;
				this.dom.removeChild(page.domItem);
			}
		}
	},
	resetMainPage: function() {
		_topPanel.toggle(false);
		var prev = this.mainPage;
		if (prev && prev.externalAdapter) {
			prev.externalAdapter.releaseMainPage({
				mainPage: prev,
				doEvent: function() {
					prev.dispose && prev.dispose();
				}
			});
		}
		this.dom.empty(this.body);
		this.diagnosesPanel && this.diagnosesPanel.clean();
	},
	onMainPageChange: function($itemPage) {
		this.mainPage && _url.history.queryPageSession.save(this.mainPage);
		syra_site.dom.hide(this.body, true);
		$itemPage.boxParent = null;
		$itemPage.layoutSlot = this.body;
		this.resetMainPage();
		this.mainPage = this.pageLoader.load($itemPage, true);
		this.mainPage.openerUrlSegments = $itemPage.openerUrlSegments;
		this.mainPage.isMainPage = true;
		this.refreshBrowserTitle(this.mainPage);
		this.lastOpenedMainPageUrlSegments = this.mainPage.openerUrlSegments;
		_url.history.addBackButton(this.mainPage);
		syra_workingCopy.showLastSavingDiagnoses(this.mainPage);
		return this.mainPage;
	},
	addDragDropManager: function() {
		window.syra_dd = _dragDropManager;
	},
	drawBox: function() {
		var self = this;
		self.addDragDropManager();
		self.domItem = self.body = document.createElement("div");
		self.body.id = "s-site-body";
		self.body.setAttribute("tabindex", "1"); // add for receive keyboard event  for shortcut
		self.layoutSlot.appendChild(self.body);
		self.productName = document.createElement("div");
		self.productName.className = 's-m-product s-m-product-font';
		self.logon(function() {
			self._onAfterLogon();
		});
	},
	_onAfterLogon: function() {
		this.layoutSlot.style.display = "";
		this.resize();
		_url.history.start();
		this.bookmarksManager && this.bookmarksManager.loadBookmarks();
	},
	getPreferences: function(key, defaultValues, target) {
		var values = this.userProfile.dataset.sitePreferences;
		if (values) {
			values = (values[this.$device] || {})[key];
		}
		if (values === undefined && this.$preferences) {
			values = this.$preferences[key];
		}
		values = values || {};
		if (defaultValues) {
			var keys = Object.keys(defaultValues);
			for (var ii = 0, jj = keys.length; ii < jj; ii++) {
				if (values[keys[ii]] === undefined) {
					values[keys[ii]] = defaultValues[keys[ii]];
				}
			}
		}
		if (target) {
			this.applyValues(target, values);
		}
		return values;
	},
	setPreferences: function(key, source, props) {
		var sitePreferences = this.userProfile.dataset.sitePreferences = this.userProfile.dataset.sitePreferences || {};
		var preferences = sitePreferences[this.$device] = sitePreferences[this.$device] || {};
		if (source || props) {
			if (props) {
				this.applyValues(preferences[key] = preferences[key] || {}, source, props);
			} else {
				preferences[key] = source;
			}
		} else {
			delete sitePreferences[this.$device][key];
		}
		this.userProfile.saveSitePreferences();
	},
	applyValues: function(target, values, keys) {
		if (target) {
			keys = keys || Object.keys(values);
			for (var ii = 0, jj = keys.length; ii < jj; ii++) {
				target[keys[ii]] = values[keys[ii]];
			}
		}
	},
	registerScreenDesigner: function(add, designer) {
		if (add) {
			if (!this.screenDesigners) {
				this.screenDesigners = [];
			} else {
				var top = this.screenDesigners && this.screenDesigners.length && this.screenDesigners[this.screenDesigners.length - 1];
				if (top) {
					top.onTopScreenDesigner(false);
				}
			}
			syra_site.dom.hide(this.header, true);
			this.screenDesigners.push(designer);
		} else {
			if (this.screenDesigners) {
				var index = this.screenDesigners.indexOf(designer);
				if (index >= 0) {
					this.screenDesigners.splice(index, 1);
				}
				if (this.screenDesigners.length == 0) {
					delete this.screenDesigners;
					syra_site.dom.hide(this.header, false);
				} else {
					var top = this.screenDesigners && this.screenDesigners.length && this.screenDesigners[this.screenDesigners.length - 1];
					if (top) {
						top.onTopScreenDesigner(true);
					}
				}
			}
		}
	},
	closePopups: function(box, event) {
		_topPanel.toggle(false, event, true);
		event && syra_dd.stop();
		_dialogManager.closePopups(box, event);
	},

	setProductName: function(name) {
		if (this.productName) {
			this.productName.textContent = name || "";
			this.bookmarksManager && this.bookmarksManager.onResize();
		}
	},
	resize: function() {
		this.layoutRect = null;
		this.autoViewMode = (window.innerWidth < 900 || (window.innerHeight > window.innerWidth)) ? "float" : "dock";
		if (this.body) {
			var bodyRect = this.dom.getBoundingClientRect(this.body);
			this.body.style.height = (this.getLayoutSize().height - bodyRect.top) + "px";
		}
		this.pageLoader.resize();
		this.bookmarksManager && this.bookmarksManager.onResize();
		this.uiLocker.ensureLock();
	},
	_renderHeader: function() {
		this.header = document.createElement("header");
		this.header.id = "s-site-header";
		this.layoutSlot.insertBefore(this.header, this.layoutSlot.firstChild);

		this.headerTop = document.createElement("div");
		this.headerTop.id = "s-site-header-top";

		this.siteHelpLink = this.loadNewItem(this.headerTop, {
			$clientId: "s-site-help-link",
			$bind: "$help",
			$category: "link",
			$noText: true,
			$skin: "s-site-help-link",
			$icon: {
				$value: "help"
			}
		});
		this.addPageDesignerOpener();

		var middleCell = document.createElement("div");
		middleCell.id = "s-site-header-middle-cell";
		this.headerTop.appendChild(middleCell);

		_trackersManager.addOpener(middleCell);

		this.setArticleId(this.layoutSlot);
		_topPanel.addOpener(middleCell);

		this.addLogoutButton();

		this.homeLink = this.loadNewItem(this.headerTop, {
			$bind: "$home",
			$category: "link",
			$noText: true,
			$skin: "s-site-home",
			$shortCutTip: "ESC G H"
		});

		this.navigationLink = this.loadNewItem(this.headerTop, {
			$bind: "$navigation",
			$category: "link",
			$noText: true,
			$skin: "s-site-navigation",
			$shortCutTip: "ESC G N"
		});

		this.searcher = _searcher.load(this.headerTop);

		this.header.appendChild(this.headerTop);

		this.headerBottom = document.createElement("div");
		this.headerBottom.id = "s-site-header-bottom";

		this.addLogo(this.headerBottom);

		this.bookmarksManager = _bookmarksManager;
		this.bookmarksManager.load();

		if (this.$isLicenceTool) {
			_domHelper.hide(this.searcher && this.searcher.domItem, true);
			this.navigationLink.setMenu({
				$isHidden: true
			});
		}

		this.header.appendChild(this.headerBottom);
	},
	isLandscape: function() {
		return window.orientation && (window.orientation == 90 || window.orientation == -90);
	},
	addPageDesignerOpener: function() {
		this.pageDesignerBtn = syra_menus.button.add({
			parent: this,
			slot: this.headerTop,
			text: syra_local.userProfile_siteAuthoringOpen,
			css: "s-page-designer-opener",
			iconOnly: true,
			btnclick: function() {
				var page;
				var site = this.parent;
				site.topPanel.toggle(false);
				if (site.mainPage && site.mainPage.isFusionPage) {
					page = site.fusionGateway.activatedBook.selectedSheet;
				} else {
					page = site.dialogManager.getTopDialogPage();
					if (page) {
						page = page._content;
					}
				}
				page = page || site.mainPage;
				if (page) {
					if (site.currentQuickEdit &&
						site.currentQuickEdit.needCloseConfirm(page, function() {
							exports.onOpenerClick();
						})) {
						return;
					}
					site.switchItemDesigner(page, true);
				}
			}
		});
		syra_menus.button.visibility(this.pageDesignerBtn, false);

	},
	switchItemDesigner: function(item, open) {
		if (item.switchDesigner) {
			if (open && !this._isCssLoaded) {
				this._isCssLoaded = true;
				this.loadStyleSheet("authoring.css");
			}
			item.switchDesigner(open);
		}
	},

	refreshMainPageSecurity: function(page) {
		if (page && !page.disposed) {
			var role = this.userProfile.getSelectedRole();
			var endpoint = this.userProfile.getSelectedEndpoint();
			if (role || endpoint) {
				role = role && role.description;
				endpoint = endpoint && endpoint.description;
				var pages = page ? [page] : this.pageLoader.getLoaded();
				for (var ii = 0, jj = pages.length; ii < jj; ii++) {
					page = pages[ii];
					if (page.securityViewSlot) {
						this.dom.empty(page.securityViewSlot);
						syra_menus.button.remove(page.securityRoleFlag);
						syra_menus.button.remove(page.securityEndpointFlag);
						if (role) {
							page.securityRoleFlag = syra_menus.button.add({
								isIndicator: true,
								parent: page,
								slot: page.securityViewSlot,
								text: role,
								css: "s-security-user",
								fontIcon: "role"
							});
						}
						if (endpoint) {
							page.securityEndpointFlag = syra_menus.button.add({
								isIndicator: true,
								parent: page,
								slot: page.securityViewSlot,
								text: endpoint,
								css: "s-security-user",
								fontIcon: "endpoint"
							});
						}
					}
				}
			}
			if (this.pageDesignerBtn && page.mainPageDesignerAccess != "unknow") {
				var show = page.mainPageDesignerAccess != "disabled" && this.userProfile.hasDesignRight();
				this.pageDesignerBtn.link.style.visibility = show ? "" : "hidden";
			}
		}
	},
	loadBox: function() {
		this.documentHeader = document.getElementsByTagName("head")[0];
		this.isDocumentRTL = document.dir === 'rtl';
		this.isSiteRegisterDisabled = true;
		this.helpCenter = _helpCenter;

		this.pageLoader = pageLoader;

		this.uiLocker = uiLocker;
		this.dom = _domHelper;
		this.layoutUpdater = _layoutUpdater;
		this.layoutValidator = _layoutValidator;
		this.deltaManager = _deltaManager;
		this.expressionMaker = expressionMaker;
		this.dialogManager = _dialogManager;

		this.filterMaker = filterMaker;

		this.localizer = localizer;
		this.topPanel = _topPanel;
		this.maximizer = _maximizer;
		window.syra_url = _url;
		window.syra_tip = _tip;
		window.syra_workingCopy = _workingCopy;
		window.syra_quality = _qualityController.create();
		window.syra_trackers = _trackersManager;
		window.syra_diagnose = _diagnoseCenter;
		window.syra_page = pageHelper;
		window.syra_menus = _menusHelper;
		window.syra_fields = _fieldHelper;
		window.syra_site = this;
		window.syra_local = locale.resources(module)();

		this.$device = this.$device || "desktop";
		if (this.isTabletDevice = document.location.pathname.indexOf("-tablet.html") >= 0) {
			this.$device = "tablet";
			this.$preferences = {
				convergenceBar: {
					$viewMode: "float",
					$isCollapsed: true
				},
				menuBar: {
					$viewMode: "dock",
					$isCollapsed: !this.isLandscape
				},
				"dash-menubar": {
					$viewMode: "dock",
					$isCollapsed: false
				}
			};
		}
		this.$prototype = this.$prototype || {};
		this.$prototype.$properties = this.$prototype.$properties || {};
		this.$item = this.$item || {};
		var $links = this.$prototype.$links = this.$prototype.$links || {};

		$links.$navigation = {
			$url: "?representation=home.$navigation&profile={$userprofile}"
		};
		$links.$home = {
			$url: "?representation=home.$landing&profile={$userprofile}"
		};
		$links.$help = {
			$url: "",
			$type: "text/html",
			$target: "help"
		};
		$links.$designerDocumentation = {
			$url: "/help/{helpLanguage}/tutorial/central-page",
			$type: "text/html",
			$target: "help"
		};
		$links.$documentation = {
			$url: "/help/{helpLanguage}/index.htm",
			$type: "text/html",
			$target: "help"
		};
		$links.$shortcuts = {
			$url: "/help/{helpLanguage}/ui-definition_keyboard-shortcuts.html",
			$type: "text/html",
			$target: "help"
		};
		$links.$legal = {
			$url: ""
		};
		$links.$support = {
			$url: "http://www.sage.com/support"
		};
		this.$navigation = {
			changeLog: {
				$title: "Change Log",
				$url: "/changelog.html",
				$type: "text/html",
				$target: "blank"
			},
			tablethome_v7: {
				$title: "Tablet Home",
				$url: "/syracuse-main/html/main-tablet.html?url=%3Frepresentation%3Dhome.%24landing"
			},
			mobileHome_v7: {
				"$title": "Mobile Home V7",
				"$description": "V7 Mobile application home page",
				"$url": "/syracuse-mobile/html/mobile.html#start"
			},
			mobileHome: {
				"$title": "Mobile Home",
				"$description": "Mobile application home page",
				"$url": "/syracuse-tablet/dist/index.html"
			},
			mobileDesktopAuthoring: {
				$title: "Mobile Desktop Personalization",
				"$description": "Mobile personalization with desktop client",
				$url: "/syracuse-main/html/main-mobile.html?url=%3Frepresentation%3Dhome.%24landing"
			},
			devsos: {
				$title: "Developer samples/SOS",
				$url: "?representation=devsos.$dev"
			},
			arrayTest: {
				$title: "UI test Pages",
				$url: "?representation=s-uitest.$test&name=array_array&category=grid"
			},
			perfmon: {
				$title: "Performance data",
				$url: "/syracuse-perfmon/public/render.html?dummy",
				$type: "text/html",
				$target: "blank"
			}
		};

		this.$maxBookmarks = 25;
		this.$legalDataUrl = "/sdata/syracuse/collaboration/syracuse/licenseDatas?representation=licenseData.$query";
		this.$legalPrototypeUrl = "/sdata/syracuse/collaboration/syracuse/$prototypes('licenseData.$query')";
		this.$iconPath = "/syracuse-ui/themes/desktop/images/";
		this.$userProfileUrl = this.$userProfileUrl || "/sdata/syracuse/collaboration/syracuse/userProfiles/$template/$workingCopies?representation=userProfile.$edit&role={$role}";
		this.$pageTemplateUrl = "/sdata/syracuse/collaboration/syracuse/pages('{endpoint}.{representation},{target},{pageview}')?role={$role}&profile={$userprofile}";
		if (this.$device == "mobile") {
			this.$pageTemplateUrl = "/sdata/syracuse/collaboration/syracuse/pages('{endpoint}.{representation},{target},{pageview},phone')?role={$role}&profile={$userprofile}";
			this.$userProfileUrl += "&device=phone";
		}

		window.syra_store = _store;
		this.widgetsLibrary = _widgetsLibrary;

		this._childPageOffset = 1;
		this.pageLoader.initialize(this);
		window.syra_controller = _controller;

		if (!this.isOfficeSite) {
			this.fusionGateway = new FusionGateway();
			this.fusionGateway.initialize();
		}
		this.externalAdapter = new ExternalAdapter();

		this.layoutSlot = document.getElementById("s-site");
		syra_site.dom.hide(this.layoutSlot, true);

		eventListener.load();


		this.initializeNewItem(this, this.$item);

		this.setArticleId(this.layoutSlot);
		RawPage.prototype.loadBox.call(this);
		return this;
	},
	addLogo: function(slot) {
		var title, href;
		var home = this.menuItems && this.menuItems.$home;
		home = home && home[0];
		if (home) {
			title = home.domItem.title;
			href = home.domItem.getAttribute("href");
		}
		var btn = syra_menus.button.add({
			parent: this,
			slot: slot,
			text: title || "",
			iconOnly: true,
			css: "s-site-header-logo",
			btnclick: function() {
				syra_menus.click.home();
			}
		});
		href && btn.link.setAttribute("href", href);

		this.brandName = document.createElement("div");
		this.brandName.className = 's-m-brand s-m-brand-font';
		this.brandName.textContent = "Sage";
		btn.link.appendChild(this.brandName);

		if (this.productName) {
			btn.link.appendChild(this.productName);
		}
	},
	addLogoutButton: function() {
		syra_menus.button.add({
			parent: this,
			slot: this.headerTop,
			text: syra_local.siteLogoutLink,
			iconOnly: true,
			css: "s-site-logout",
			btnclick: function() {
				syra_site.logout();
			}
		});
	},
	onUncaughtError: function(errorMsg, url, lineNumber, error) {
		var self = this;
		if (self.mainPage) {
			var options = {
				errorMsg: errorMsg,
				url: url,
				lineNumber: lineNumber,
				error: error,
				doEvent: function() {
					// diagnose component will be improved, with specific msg and stack properties
					var $diagnose = {};
					$diagnose.$message = options.errorMsg;
					$diagnose.$message += error ? "\nStack:\n" + error.stack : "";
					$diagnose.$severity = "error"; //"alert";
					syra_diagnose.showDiagnoses({
						$diagnoses: [$diagnose]
					});
				}
			};
			if (self.mainPage.externalAdapter)
				self.mainPage.externalAdapter.onUncaughtError(options);
		} else {
			syra_diagnose.showDiagnoses({
				$diagnoses: [{
					$severity: "fatal",
					$message: errorMsg
				}]
			});
		}
	},
	onError: function(error) {
		this.onUncaughtError(error.message, error.fileName, error.lineNumber, error);
	},
	getLayoutSize: function() {
		if (!this.layoutRect) {
			this.layoutRect = this.dom.getBoundingClientRect(this.layoutSlot);
		}
		return this.layoutRect;
	},
	refreshBrowserTitle: function(page) {
		if (page == this.mainPage) {
			var title = page.getTitle();
			if (title) {
				if (this.$isSecurityTitleVisible !== false) {
					var role = syra_site.userProfile.getSelectedRole();
					var endpoint = syra_site.userProfile.getSelectedEndpoint();
					if (!this.tabBrowserTitles) {
						this.tabBrowserTitles = {
							$edit: syra_local.tab_browser_facet_edit,
							$details: syra_local.tab_browser_facet_details,
						};
					}
					var facet = this.tabBrowserTitles[page.$facet];
					if (facet) {
						title += " - " + facet + " ";
					}
					if (role) {
						title += " (" + role.description + ")";
					}
					if (endpoint) {
						title += " (" + endpoint.description + ")";
					}
				}
				document.title = title;
			}
		}
	},
	onItemInOut: function(onEnter, event, target) {
		_topPanel.onItemInOut(onEnter, event, target);
	},
	ensureRecordUniqueSelection: function(list, onSelect) {
		var record;
		while (list && !list.isArrayField) {
			if (list.isRecordArticle) {
				record = list;
			}
			list = list.articleParent;
		}
		if (list && !list.isArrayField) {
			list = null;
		}
		if (this._selectedList && this._selectedList.selector) {
			if (list != this._selectedList || (record && !record.$isSelected && !this._selectedList.selector.isMulti)) {
				this._selectedList.selector.unSelectAll();
			}
		}
		if (onSelect) {
			this._selectedList = list;
		}
	},
	disposeObject: function(obj) {
		if (obj) {
			var keys = Object.keys(obj);
			for (var ii = 0, jj = keys.length; ii < jj; ii++) {
				obj[keys[ii]] = null;
			}
			obj.disposed = true;
		}
	}
});
