// syracuse-ui/lib/site/widgetsLibrary.js
"use strict";
var MenuItem = require('syracuse-ui/lib/menus/menuItem').MenuItem;
var MenusBox = require('syracuse-ui/lib/menus/menusBox').MenusBox;
var SectionBlock = require('syracuse-ui/lib/article/box').SectionBlock;
var SearchFacets = require('syracuse-ui/lib/article/searchFacets').SearchFacets;
var VignetteField = require('syracuse-ui/lib/field/vignette/vignetteField').VignetteField;
var flashChart = require('syracuse-ui/lib/field/chart/flashChart');
var highCharts = require('syracuse-ui/lib/field/chart/highCharts');
//var fusionChartsXT = require('syracuse-ui/lib/field/chart/fusionChartsXT');
var testFactory = require('syracuse-ui/lib/tests/testFactory');
var alphaNumField = require('syracuse-ui/lib/desktop/field/alphaNumField');
var Field = require('syracuse-ui/lib/field/field').Field;
var Slot = require('syracuse-ui/lib/article/slot').Slot;

var DateTimeField = require('syracuse-ui/lib/field/datetimeField').DateTimeField;
var NumericField = require('syracuse-ui/lib/field/numericField').NumericField;

exports.fieldTypes = {
    "application/x-icon": require('syracuse-ui/lib/field/iconField').IconField,
    "application/x-choice": require('syracuse-ui/lib/field/choice/choiceField').ChoiceField,
    "application/x-date": DateTimeField,
    "application/x-time": DateTimeField,
    "application/x-datetime": DateTimeField,
    "application/x-binary": require('syracuse-ui/lib/field/binary/fileField').FileField,
    "image": require('syracuse-ui/lib/field/binary/imageField').ImageField,
    "application/x-boolean": require('syracuse-ui/lib/field/booleanField').BooleanField,
    "application/x-integer": NumericField,
    "application/x-decimal": NumericField,
    "application/x-real": NumericField,
    "application/x-password": require('syracuse-ui/lib/field/passwordField').PasswordField,
    "application/x-string$text": Field,
    "application/x-string$phone": alphaNumField.PhoneField,
    "application/x-string$email": alphaNumField.EmailField,
    "application/x-string$url": alphaNumField.UrlField,
    "application/x-array": require("syracuse-ui/lib/field/collection/list").DesktopList,
    "application/x-reference": require('syracuse-ui/lib/field/referenceField').ReferenceField,
    "application/x-cube": require('syracuse-ui/lib/field/cube/cubeField'),
    "application/x-vignette": VignetteField,
    "application/x-portlet": VignetteField,
    "application/x-graph$d3": require('syracuse-ui/lib/field/chart/d3Graph'),
    "application/x-tag-cloud$d3": require('syracuse-ui/lib/field/chart/d3TagCloud'),
    "application/x-calendar": require('syracuse-ui/lib/field/schedule/calendar').Calendar,
    "application/x-gantt": require('syracuse-ui/lib/field/schedule/gantt').Gantt,
    "application/x-process": require('syracuse-ui/lib/desktop/field/process/visualProcess').VisualProcess,
    "text/rtf": require('syracuse-ui/lib/field/editor/htmlEditor').HtmlEditor
};

var _layout = require("syracuse-ui/lib/article/layouts/layout").Layout;
exports.layouts = {
    columns: require("syracuse-ui/lib/article/layouts/columnsLayout").ColumnsLayout,
    side: _layout,
    row: _layout,
    stack: _layout,
    tabs: require("syracuse-ui/lib/article/layouts/tabsLayout").TabsLayout,
    composite: _layout
};

exports.defaultPageCategory = require('syracuse-ui/lib/article/page').DesktopPage;
exports.pageCategories = {
    fusionpage: require('syracuse-ui/lib/fusion/article/page').FusionPage
};

exports.categories = {
    slot: {
        create: function(){
            return new Slot();
        }
    },
    tests: {
        create: function($item){
            return new testFactory[$item.$format]();
        }
    },
    searchFacets: {
        create: function(){
            return new SearchFacets();
        }
    },
    section: {
        create: function($item){
            var item = new SectionBlock();
            item.$authoringLevel = "section";
            return item;
        }
    },
    block: {
        create: function(){
            var item = new SectionBlock();
            item.$authoringLevel = "block";
            return item;
        }
    },
    links: {
        create: function(){
            return new MenusBox();
        }
    },
    link: {
        create: function(){
            return new MenuItem();
        }
    },
    fusionBar: {
        create: function($item, page){
            if (page.appendFusionBar) {
                page.appendFusionBar($item);
            }
            return null;
        }
    },
    tabularChart: {
        types: {
            x3flash: flashChart.TabularChart,
            highCharts: highCharts.TabularChart
            /*,
             fusionChartsXT: fusionChartsXT.TabularChart*/
        },
        create: function($item){
            var itemClass = this.types[$item.$chartType || "highCharts"];
            // var itemClass = this.types[$item.$chartType || "x3flash"];
            return new itemClass();
        }
    },
    field: {
        create: function($item, boxParent){
            var article = boxParent.getArticle();
            var $field = $item.$field || article.get$Properties()[$item.$bind];
            var field;
            if ($field) {
                var $class = $field.$type;
                var fieldClass;
                switch ($class) {
                    case "application/x-tag-cloud":
                        var $format = $field.$format || $item.$format || "d3.tagCloud";
                        var parts = $format.split(".");
                        fieldClass = exports.fieldTypes[$class + "$" + parts[0]][parts[1]];
                        break;
                    case "application/x-graph":
                        var $format = $field.$format || $item.$format || "d3.nodeChart";
                        var parts = $format.split(".");
                        fieldClass = exports.fieldTypes[$class + "$" + parts[0]][parts[1]];
                        break;
                    case "application/x-string":
                        $class += ($field.$format || "$text");
                        break;
                }
                /*if (!article.getPage()) {
                 debugger;
                 }*/
                field = article.getPage().externalAdapter.createField($field, $item, boxParent, $class);
                if (!field) {
                    fieldClass = fieldClass || exports.fieldTypes[$class];
                    if (fieldClass) {
                        (field = new fieldClass()).$field = $field;
                    }
                }
                if (!field) {
                    console.log("field type unknow:" + $item.$bind + ($field.$type || ""));
                }
            }
            else {
                console.log($item.$bind + " is defined in Layout but not in proto");
            }
            return field;
        }
    }
};
