// syracuse-ui/lib/field/datetimeSelectors/dateSelector.js
"use strict";
//http://www.openweb.eu.org/articles/formulaire_accessible
var helpers = require('syracuse-core/lib/helpers');
var _dateApi = require('syracuse-core/lib/types/date');
var MonthChoice = require('syracuse-ui/lib/field/datetimeSelectors/monthChoice').MonthChoice;
var YearChoice = require('syracuse-ui/lib/field/datetimeSelectors/yearChoice').YearChoice;

function DateSelector() {

}

exports.DateSelector = helpers.defineClass(DateSelector, null, {
	onTodayClick: function() {
		this._currentDate = _dateApi.today();
		this.field.setPickerValue(_dateApi.fromInternalValue(this._currentDate._value).toString(this.field.localFormat), this._currentDate._value);
	},
	onDayClick: function(picker) {
		this.field.setPickerValue(_dateApi.fromInternalValue(picker.syraValue).toString(this.field.localFormat), picker.syraValue);
	},
	onQuickItemClick: function(picker, event) {
		var isBegin = syra_site.dom.getChildIndex(event.target) == 0;
		this._currentDate = this._currentDate[(isBegin ? "begOf" : "endOf") + picker.syraInterval]((picker.syraInterval == "Week") ? 1 : undefined);
		this._drawBody();
		this.quickList.style.display = "none";
	},
	onChangePeriode: function(picker) {
		if (picker.syraPeriod == "month") {
			this._currentDate = this._currentDate.addMonths(picker.syraIsPrev ? -1 : 1);
		} else
		if (picker.syraPeriod == "year") {
			this._currentDate = this._currentDate.addYears(picker.syraIsPrev ? -1 : 1);
		} else {
			this._currentDate = this._currentDate.addDays(picker.syraIsPrev ? -7 : 7);
		}
		this._drawBody();
	},
	onQuickClick: function() {
		if (!this.quickList) {
			this.quickList = document.createElement("nav");
			this.quickList.style.display = "none";
			this.quickList.className = "s-calendar-quick-list";
			var intervals = ["Year", "Quarter", "Month", "Week"];
			for (var ii = 0, jj = intervals.length; ii < jj; ii++) {
				var row = document.createElement("div");
				row.className = "s-calendar-quick";
				row.syraInterval = intervals[ii];
				row.syraOnClick = "onQuickItemClick";
				row.innerHTML = syra_site.formatHTMLText("<a class='s-calendar-quick-link'>" + //
					this.field.localize.fdpBegin + //
					"</a>/<a class='s-calendar-quick-link'>" + //
					this.field.localize.fdpEnd + "</a>" + //
					this.field.localize["fdpIntervalOf" + intervals[ii]], false);
				this.quickList.appendChild(row);
			}
			this.domItem.appendChild(this.quickList);
		}
		if (this.quickList.style.display == "") {
			this.quickList.style.display = "none";
		} else {
			this.quickList.style.display = "";
			$(this.quickList).position({
				my: "left bottom",
				at: "right bottom",
				of: $(this._quick)
			});
		}
	},
	create: function(field, options) {
		repository[this.id = (field.id + "dateselector")] = this;
		this.domItem = document.createElement("div");
		this.domItem.className = "s-calendar";
		this.domItem.style.display = "none";
		field.domItem.appendChild(this.domItem);
		this.domItem.syraTool = this.id;
		this.domItem.syraarticle = field.page.id;
		if (options && options.internalValue) {
			this._selectedDate = isNaN(options.internalValue) ? _dateApi.today() : _dateApi.fromInternalValue(options.internalValue);
			this._currentDate = _dateApi.fromInternalValue(this._selectedDate._value);
		} else {
			var value = field.getInputValue() || "";
			this._selectedDate = value ? _dateApi.parse(value, field.localFormat) : _dateApi.today();
			this._currentDate = _dateApi.fromInternalValue(this._selectedDate._value);
		}
		this.field = field;
		this._table = document.createElement("table");
		this._table.setAttribute("cellspacing", 0);
		this._table.className = "s-calendar-content";
		this._table.appendChild(this._appendHead());
		this._table.appendChild(this.body = document.createElement("tbody"));
		this._table.appendChild(this._appendFoot());
		this.domItem.appendChild(this._table);

		this._drawBody();
	},
	remove: function() {
		syra_site.dom.removeChild(this.domItem);
		this.dispose();
	},
	_appendHead: function() {
		var head = document.createElement("thead");
		var row = document.createElement("tr");
		var cell = document.createElement("th");
		cell.setAttribute("colspan", 8);
		cell.className = "s-calendar-month-year";

		// prev year link
		var link = document.createElement("a");
		link.syraOnClick = "onChangePeriode";
		link.syraIsPrev = true;
		link.syraPeriod = "year";
		link.className = "s-calendar-prev s-calendar-prev-year";
		cell.appendChild(link);

		// prev month link
		link = document.createElement("a");
		link.syraOnClick = "onChangePeriode";
		link.syraPeriod = "month";
		link.syraIsPrev = true;
		link.className = "s-calendar-prev s-calendar-prev-month";
		cell.appendChild(link);

		this._monthLink = document.createElement("a");
		this._monthLink.className = "s-calendar-month-year-link";
		cell.appendChild(this._monthLink).syraOnClick = "onMonthClick";

		this._yearLink = document.createElement("a");
		this._yearLink.className = "s-calendar-month-year-link";
		cell.appendChild(this._yearLink).syraOnClick = "onYearClick";

		link = document.createElement("a");
		link.syraOnClick = "onChangePeriode";
		link.syraPeriod = "month";
		link.className = "s-calendar-next s-calendar-next-month";
		cell.appendChild(link);

		link = document.createElement("a");
		link.syraOnClick = "onChangePeriode";
		link.syraPeriod = "year";
		link.className = "s-calendar-next s-calendar-next-year";
		cell.appendChild(link);

		row.appendChild(cell);
		head.appendChild(row);

		row = document.createElement("tr");
		var cell = document.createElement("th");
		cell.className = "s-calendar-week-day";
		row.appendChild(cell);

		var days = [1, 2, 3, 4, 5, 6, 0];
		for (var ii = 0, jj = days.length; ii < jj; ii++) {
			var cell = document.createElement("th");
			cell.className = "s-calendar-week-day";
			cell.title = _dateApi.dayName(days[ii]);
			cell.textContent = _dateApi.dayName(days[ii], true);
			row.appendChild(cell);
		}
		head.appendChild(row);
		return head;
	},
	_appendFoot: function() {
		var row = document.createElement("tr");
		var cell = document.createElement("td");
		cell.setAttribute("colspan", 3);
		cell.className = "s-calendar-foot-week";

		var link = document.createElement("a");
		link.syraOnClick = "onChangePeriode";
		link.syraPeriod = "week";
		link.className = "s-calendar-prev";
		cell.appendChild(link);

		var label = document.createElement("label");
		label.className = "s-calendar-foot-week-title";
		label.textContent = this.field.localize.fdpWeek;
		cell.appendChild(label);

		this._weekNumber = document.createElement("label");
		this._weekNumber.className = "s-calendar-foot-week-title-num";
		cell.appendChild(this._weekNumber);

		link = document.createElement("a");
		link.syraOnClick = "onChangePeriode";
		link.syraPeriod = "week";
		link.className = "s-calendar-next";
		cell.appendChild(link);
		row.appendChild(cell);

		cell = document.createElement("td");
		cell.setAttribute("colspan", 3);
		cell.className = "s-calendar-foot-today";
		link = document.createElement("a");
		link.className = "s-calendar-today-link";
		link.syraOnClick = "onTodayClick";
		link.textContent = this.field.localize.fdpToday;
		cell.appendChild(link);
		row.appendChild(cell);

		cell = document.createElement("td");
		cell.setAttribute("colspan", 2);
		cell.className = "s-calendar-foot-more";
		this._quick = document.createElement("a");
		this._quick.className = "s-calendar-quick-btn";
		this._quick.syraOnClick = "onQuickClick";
		cell.appendChild(this._quick);
		row.appendChild(cell);

		var foot = document.createElement("tfoot");
		foot.appendChild(row);
		return foot;
	},

	_drawBody: function() {
		var curMonth = this._currentDate.month;
		var month = _dateApi.monthName(curMonth);

		this._monthLink.textContent = month;
		this._yearLink.textContent = this._currentDate.year;

		syra_site.dom.empty(this.body);

		var curDate = _dateApi.fromInternalValue(this._currentDate._value);
		var begOfMonth = curDate = curDate.begOfMonth();
		curDate = curDate.begOfWeek(1);
		for (var weekRow = 0; weekRow < 6; weekRow++) {
			var row = document.createElement("tr");
			var weekDay = (weekRow == 0) ? begOfMonth : curDate;
			var cell = document.createElement("td");
			cell.className = "s-calendar-week-num";
			row.appendChild(cell);
			cell.textContent = weekDay.week;
			for (var day = 0; day < 7; day++) {
				var cell = document.createElement("td");
				cell.className = "s-calendar-day-link";
				cell.syraOnClick = "onDayClick";
				cell.syraValue = curDate._value;
				var link = document.createElement("a");
				link.textContent = curDate.day;
				if (curMonth != curDate.month) {
					cell.className += " s-calendar-other-month";
				}
				if (this._currentDate.equals(curDate)) {
					cell.className += " s-calendar-select";
					link.className = "s-calendar-select";
				}
				cell.appendChild(link);
				row.appendChild(cell);
				curDate = curDate.addDays(1);
			}
			this.body.appendChild(row);
		}
		this._weekNumber.textContent = this._currentDate.week;
	},
	onMonthClick: function() {
		var self = this;
		if (!self._month) {
			self._month = new MonthChoice();
			self._month.create(self.field, self.domItem, self._currentDate, function(newDate) {
				self._currentDate = newDate;
				setTimeout(function() {
					self._drawBody();
					self.onMonthClick();
				}, 200);
				return true;
			});
			self._month.toggle(true);
		} else {
			self._month.toggle(false);
			self._month.dispose();
			delete self._month;
		}
	},
	onYearClick: function() {
		var self = this;
		if (!self._year) {
			self._year = new YearChoice(self.field);
			self._year.create(self.field, self.domItem, self._currentDate, function(newDate) {
				self._currentDate = newDate;
				setTimeout(function() {
					self._drawBody();
					self.onYearClick();
				}, 200);
				return true;
			});
			self._year.toggle(true);
		} else {
			self._year.toggle(false);
			self._year.dispose();
			delete self._year;
		}
	},
	dispose: function() {
		delete repository[this.id];
		this.domItem = this._selectedDate = this._currentDate = this.field = this._table = this._quick = this.quickList = null;
	}
});
