// syracuse-ui/lib/field/datetimeSelectors/yearChoice.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var _dateApi = require('syracuse-core/lib/types/date');

function YearChoice() {

}

exports.YearChoice = helpers.defineClass(YearChoice, null, {
	create: function(field, parent, dateValue, close) {
		syra_map[this.id = (field.id + "yearchoice")] = this;
		this.parent = parent;
		this.domItem = document.createElement("nav");
		this.domItem.className = "s-calendar-quick-list";
		this.domItem.syraTool = this.id;
		this.domItem.syraarticle = field.page.id;
		this._close = close;
		this._dateValue = dateValue;
		this._selectedYear = this._dateValue.year;
		var tbody = document.createElement("tbody");
		this.rows = [];
		for (var ii = 0; ii < 6; ii++) {
			var row = document.createElement("tr");
			if (ii == 0) {
				this.addPrevNextCell(row, true);
				this.addPrevNextCell(row, false);
			} else {
				var cuYear = this._dateValue.year + ii - 4;
				row.appendChild(this.drawCell(cuYear, cuYear));
				row.appendChild(this.drawCell(cuYear + 5, cuYear + 5));
			}
			tbody.appendChild(row);
			this.rows.push(row);
		}
		this.domItem.appendChild(tbody);
		this.domItem.style.display = "none";
		this.parent.insertBefore(this.domItem, this.parent.firstChild);
	},
	addPrevNextCell: function(row, isPrev) {
		var td = document.createElement("td");
		td.className = "s-calendar-my-link-cell";
		var btn = syra_menus.addIconButton(isPrev ? syra_local.fdpPrevYear : syra_local.fdpNextYear, isPrev ? "s-calendar-prev" : "s-calendar-next", "onChangeYear");
		btn.syraIsPrev = isPrev;
		td.appendChild(btn);
		row.appendChild(td);
	},
	drawCell: function(value, text) {
		var td = document.createElement("td");
		td.className = "s-calendar-my-item";
		td.syraOnClick = "onSelectYear";
		td.syraValue = value;
		td.textContent = text;
		if (value == this._selectedYear) {
			td.className += " s-calendar-select";
			this.selectedCell = td;
		}
		return td;
	},
	toggle: function(show) {
		if (show) {
			this.domItem.style.display = "";
			$(this.domItem).position({
				my: "left top",
				at: "right top",
				of: $(this.parent)
			});
		} else {
			this.domItem.style.display = "none";
		}
	},
	_refreshYearCell: function(td, step) {
		td.syraValue += step;
		td.textContent = td.syraValue;
		if (td.syraValue == this._selectedYear) {
			td.className += " s-calendar-select";
			this.selectedCell = td;
		}
	},
	onChangeYear: function(event, btn) {
		var step = btn.syraIsPrev ? (-10) : 10;
		for (var ii = 1, jj = this.rows.length; ii < jj; ii++) {
			var nodes = this.rows[ii].childNodes;
			this._refreshYearCell(nodes[0], step);
			this._refreshYearCell(nodes[1], step);
		}
	},
	onSelectYear: function(event, btn) {
		if (this.selectedCell) {
			this.selectedCell.className = "s-calendar-my-item";
		}
		(this.selectedCell = btn).className = "s-calendar-my-item s-calendar-select";
		this._selectedYear = btn.syraValue;
		var year = this._selectedYear ? this._selectedYear : this._dateValue.year;
		this._dateValue = _dateApi.fromJsDate(new Date(year, this._dateValue.month, this._dateValue.day));
		this._close(this._dateValue);
	},
	dispose: function() {
		delete syra_map[this.id];
		this.rows = this.parent = this.domItem = this._dateValue = this._close = this.selectedCell = null;
	}

});
