// hdb/lib-cov/protocol/reply/Part.js
/* automatically generated by JSCoverage - do not edit */
if (typeof _$jscoverage === 'undefined') _$jscoverage = {};
if (! _$jscoverage['protocol/reply/Part.js']) {
  _$jscoverage['protocol/reply/Part.js'] = [];
  _$jscoverage['protocol/reply/Part.js'][14] = 0;
  _$jscoverage['protocol/reply/Part.js'][16] = 0;
  _$jscoverage['protocol/reply/Part.js'][17] = 0;
  _$jscoverage['protocol/reply/Part.js'][18] = 0;
  _$jscoverage['protocol/reply/Part.js'][20] = 0;
  _$jscoverage['protocol/reply/Part.js'][21] = 0;
  _$jscoverage['protocol/reply/Part.js'][23] = 0;
  _$jscoverage['protocol/reply/Part.js'][25] = 0;
  _$jscoverage['protocol/reply/Part.js'][26] = 0;
  _$jscoverage['protocol/reply/Part.js'][27] = 0;
  _$jscoverage['protocol/reply/Part.js'][28] = 0;
  _$jscoverage['protocol/reply/Part.js'][29] = 0;
  _$jscoverage['protocol/reply/Part.js'][30] = 0;
  _$jscoverage['protocol/reply/Part.js'][32] = 0;
  _$jscoverage['protocol/reply/Part.js'][36] = 0;
  _$jscoverage['protocol/reply/Part.js'][38] = 0;
  _$jscoverage['protocol/reply/Part.js'][39] = 0;
  _$jscoverage['protocol/reply/Part.js'][40] = 0;
  _$jscoverage['protocol/reply/Part.js'][42] = 0;
  _$jscoverage['protocol/reply/Part.js'][46] = 0;
  _$jscoverage['protocol/reply/Part.js'][47] = 0;
  _$jscoverage['protocol/reply/Part.js'][49] = 0;
  _$jscoverage['protocol/reply/Part.js'][50] = 0;
  _$jscoverage['protocol/reply/Part.js'][51] = 0;
  _$jscoverage['protocol/reply/Part.js'][52] = 0;
  _$jscoverage['protocol/reply/Part.js'][55] = 0;
  _$jscoverage['protocol/reply/Part.js'][57] = 0;
  _$jscoverage['protocol/reply/Part.js'][59] = 0;
  _$jscoverage['protocol/reply/Part.js'][60] = 0;
  _$jscoverage['protocol/reply/Part.js'][61] = 0;
  _$jscoverage['protocol/reply/Part.js'][62] = 0;
  _$jscoverage['protocol/reply/Part.js'][63] = 0;
  _$jscoverage['protocol/reply/Part.js'][65] = 0;
  _$jscoverage['protocol/reply/Part.js'][66] = 0;
  _$jscoverage['protocol/reply/Part.js'][67] = 0;
  _$jscoverage['protocol/reply/Part.js'][68] = 0;
  _$jscoverage['protocol/reply/Part.js'][69] = 0;
  _$jscoverage['protocol/reply/Part.js'][70] = 0;
  _$jscoverage['protocol/reply/Part.js'][72] = 0;
  _$jscoverage['protocol/reply/Part.js'][75] = 0;
  _$jscoverage['protocol/reply/Part.js'][76] = 0;
  _$jscoverage['protocol/reply/Part.js'][77] = 0;
  _$jscoverage['protocol/reply/Part.js'][79] = 0;
  _$jscoverage['protocol/reply/Part.js'][81] = 0;
  _$jscoverage['protocol/reply/Part.js'][83] = 0;
  _$jscoverage['protocol/reply/Part.js'][85] = 0;
  _$jscoverage['protocol/reply/Part.js'][87] = 0;
  _$jscoverage['protocol/reply/Part.js'][89] = 0;
  _$jscoverage['protocol/reply/Part.js'][90] = 0;
  _$jscoverage['protocol/reply/Part.js'][91] = 0;
  _$jscoverage['protocol/reply/Part.js'][92] = 0;
  _$jscoverage['protocol/reply/Part.js'][94] = 0;
  _$jscoverage['protocol/reply/Part.js'][97] = 0;
  _$jscoverage['protocol/reply/Part.js'][98] = 0;
  _$jscoverage['protocol/reply/Part.js'][99] = 0;
  _$jscoverage['protocol/reply/Part.js'][100] = 0;
  _$jscoverage['protocol/reply/Part.js'][101] = 0;
  _$jscoverage['protocol/reply/Part.js'][102] = 0;
  _$jscoverage['protocol/reply/Part.js'][103] = 0;
  _$jscoverage['protocol/reply/Part.js'][104] = 0;
  _$jscoverage['protocol/reply/Part.js'][105] = 0;
  _$jscoverage['protocol/reply/Part.js'][106] = 0;
  _$jscoverage['protocol/reply/Part.js'][107] = 0;
  _$jscoverage['protocol/reply/Part.js'][108] = 0;
  _$jscoverage['protocol/reply/Part.js'][109] = 0;
  _$jscoverage['protocol/reply/Part.js'][110] = 0;
  _$jscoverage['protocol/reply/Part.js'][111] = 0;
  _$jscoverage['protocol/reply/Part.js'][112] = 0;
  _$jscoverage['protocol/reply/Part.js'][113] = 0;
  _$jscoverage['protocol/reply/Part.js'][114] = 0;
  _$jscoverage['protocol/reply/Part.js'][115] = 0;
  _$jscoverage['protocol/reply/Part.js'][116] = 0;
  _$jscoverage['protocol/reply/Part.js'][117] = 0;
  _$jscoverage['protocol/reply/Part.js'][119] = 0;
  _$jscoverage['protocol/reply/Part.js'][120] = 0;
  _$jscoverage['protocol/reply/Part.js'][122] = 0;
  _$jscoverage['protocol/reply/Part.js'][124] = 0;
  _$jscoverage['protocol/reply/Part.js'][125] = 0;
}
_$jscoverage['protocol/reply/Part.js'][14]++;
"use strict";
_$jscoverage['protocol/reply/Part.js'][16]++;
var util = require("../../util");
_$jscoverage['protocol/reply/Part.js'][17]++;
var common = require("../common");
_$jscoverage['protocol/reply/Part.js'][18]++;
var PartKind = common.PartKind;
_$jscoverage['protocol/reply/Part.js'][20]++;
var PART_HEADER_LENGTH = common.PART_HEADER_LENGTH;
_$jscoverage['protocol/reply/Part.js'][21]++;
var BIG_ARGUMENT_COUNT_INDICATOR = -1;
_$jscoverage['protocol/reply/Part.js'][23]++;
module.exports = Part;
_$jscoverage['protocol/reply/Part.js'][25]++;
function Part(kind, attributes, argumentCount, buffer, encoding) {
  _$jscoverage['protocol/reply/Part.js'][26]++;
  this.kind = kind || PartKind.NIL;
  _$jscoverage['protocol/reply/Part.js'][27]++;
  this.attributes = attributes || 0;
  _$jscoverage['protocol/reply/Part.js'][28]++;
  this.argumentCount = argumentCount || 0;
  _$jscoverage['protocol/reply/Part.js'][29]++;
  if (util.isString(buffer)) {
    _$jscoverage['protocol/reply/Part.js'][30]++;
    this.buffer = new Buffer(buffer, encoding || "hex");
  }
  else {
    _$jscoverage['protocol/reply/Part.js'][32]++;
    this.buffer = buffer;
  }
}
_$jscoverage['protocol/reply/Part.js'][36]++;
Object.defineProperty(Part.prototype, "byteLength", {get: (function getByteLength() {
  _$jscoverage['protocol/reply/Part.js'][38]++;
  var byteLength = PART_HEADER_LENGTH;
  _$jscoverage['protocol/reply/Part.js'][39]++;
  if (Buffer.isBuffer(this.buffer)) {
    _$jscoverage['protocol/reply/Part.js'][40]++;
    byteLength += util.alignLength(this.buffer.length, 8);
  }
  _$jscoverage['protocol/reply/Part.js'][42]++;
  return byteLength;
})});
_$jscoverage['protocol/reply/Part.js'][46]++;
Part.create = createPart;
_$jscoverage['protocol/reply/Part.js'][47]++;
Part.read = readPart;
_$jscoverage['protocol/reply/Part.js'][49]++;
function createPart(buffer, offset) {
  _$jscoverage['protocol/reply/Part.js'][50]++;
  var part = new Part();
  _$jscoverage['protocol/reply/Part.js'][51]++;
  readPart.call(part, buffer, offset);
  _$jscoverage['protocol/reply/Part.js'][52]++;
  return part;
}
_$jscoverage['protocol/reply/Part.js'][55]++;
function readPart(buffer, offset) {
  _$jscoverage['protocol/reply/Part.js'][57]++;
  offset = offset || 0;
  _$jscoverage['protocol/reply/Part.js'][59]++;
  this.kind = buffer[offset];
  _$jscoverage['protocol/reply/Part.js'][60]++;
  this.attributes = buffer[offset + 1];
  _$jscoverage['protocol/reply/Part.js'][61]++;
  this.argumentCount = buffer.readInt16LE(offset + 2);
  _$jscoverage['protocol/reply/Part.js'][62]++;
  if (this.argumentCount === BIG_ARGUMENT_COUNT_INDICATOR) {
    _$jscoverage['protocol/reply/Part.js'][63]++;
    this.argumentCount = buffer.readInt32LE(offset + 4);
  }
  _$jscoverage['protocol/reply/Part.js'][65]++;
  var length = buffer.readInt32LE(offset + 8);
  _$jscoverage['protocol/reply/Part.js'][66]++;
  offset += PART_HEADER_LENGTH;
  _$jscoverage['protocol/reply/Part.js'][67]++;
  if (length > 0) {
    _$jscoverage['protocol/reply/Part.js'][68]++;
    this.buffer = new Buffer(length);
    _$jscoverage['protocol/reply/Part.js'][69]++;
    buffer.copy(this.buffer, 0, offset, offset + length);
    _$jscoverage['protocol/reply/Part.js'][70]++;
    offset += util.alignLength(length, 8);
  }
  _$jscoverage['protocol/reply/Part.js'][72]++;
  return offset;
}
_$jscoverage['protocol/reply/Part.js'][75]++;
Part.prototype.toBuffer = (function toBuffer(size) {
  _$jscoverage['protocol/reply/Part.js'][76]++;
  var byteLength = util.alignLength(this.buffer.length, 8);
  _$jscoverage['protocol/reply/Part.js'][77]++;
  var buffer = new Buffer(PART_HEADER_LENGTH + byteLength);
  _$jscoverage['protocol/reply/Part.js'][79]++;
  buffer.writeInt8(this.kind, 0);
  _$jscoverage['protocol/reply/Part.js'][81]++;
  buffer.writeInt8(this.attributes, 1);
  _$jscoverage['protocol/reply/Part.js'][83]++;
  buffer.writeInt16LE(this.argumentCount, 2);
  _$jscoverage['protocol/reply/Part.js'][85]++;
  buffer.writeInt32LE(0, 4);
  _$jscoverage['protocol/reply/Part.js'][87]++;
  buffer.writeInt32LE(this.buffer.length, 8);
  _$jscoverage['protocol/reply/Part.js'][89]++;
  buffer.writeInt32LE(size, 12);
  _$jscoverage['protocol/reply/Part.js'][90]++;
  this.buffer.copy(buffer, PART_HEADER_LENGTH);
  _$jscoverage['protocol/reply/Part.js'][91]++;
  if (this.buffer.length < byteLength) {
    _$jscoverage['protocol/reply/Part.js'][92]++;
    buffer.fill(0, PART_HEADER_LENGTH + this.buffer.length);
  }
  _$jscoverage['protocol/reply/Part.js'][94]++;
  return buffer;
});
_$jscoverage['protocol/reply/Part.js'][97]++;
Part.prototype.inspect = (function inspect(options) {
  _$jscoverage['protocol/reply/Part.js'][98]++;
  var lines = [];
  _$jscoverage['protocol/reply/Part.js'][99]++;
  options = options || {};
  _$jscoverage['protocol/reply/Part.js'][100]++;
  var offset = new Buffer(options.indentOffset || 0);
  _$jscoverage['protocol/reply/Part.js'][101]++;
  offset.fill(32);
  _$jscoverage['protocol/reply/Part.js'][102]++;
  offset = offset.toString("ascii");
  _$jscoverage['protocol/reply/Part.js'][103]++;
  var kindName = common.PartKindName[this.kind];
  _$jscoverage['protocol/reply/Part.js'][104]++;
  lines.push(offset + "{\n");
  _$jscoverage['protocol/reply/Part.js'][105]++;
  lines.push(offset + "  kind: PartKind." + kindName + ",\n");
  _$jscoverage['protocol/reply/Part.js'][106]++;
  lines.push(offset + "  argumentCount: " + this.argumentCount + ",\n");
  _$jscoverage['protocol/reply/Part.js'][107]++;
  lines.push(offset + "  attributes: " + this.attributes + ",\n");
  _$jscoverage['protocol/reply/Part.js'][108]++;
  if (Buffer.isBuffer(this.buffer)) {
    _$jscoverage['protocol/reply/Part.js'][109]++;
    var length = this.buffer.length;
    _$jscoverage['protocol/reply/Part.js'][110]++;
    var start, end, chunk;
    _$jscoverage['protocol/reply/Part.js'][111]++;
    start = 0;
    _$jscoverage['protocol/reply/Part.js'][112]++;
    var hexstr = [];
    _$jscoverage['protocol/reply/Part.js'][113]++;
    while (start < length) {
      _$jscoverage['protocol/reply/Part.js'][114]++;
      end = start + 32 > length? length: start + 32;
      _$jscoverage['protocol/reply/Part.js'][115]++;
      chunk = this.buffer.toString("hex", start, end);
      _$jscoverage['protocol/reply/Part.js'][116]++;
      hexstr.push(offset + "    '" + chunk + "'");
      _$jscoverage['protocol/reply/Part.js'][117]++;
      start = end;
}
    _$jscoverage['protocol/reply/Part.js'][119]++;
    lines.push(offset + "  buffer: new Buffer(\n");
    _$jscoverage['protocol/reply/Part.js'][120]++;
    lines.push(hexstr.join(" +\n") + ", 'hex')\n");
  }
  else {
    _$jscoverage['protocol/reply/Part.js'][122]++;
    lines.push(offset + "  buffer: null\n");
  }
  _$jscoverage['protocol/reply/Part.js'][124]++;
  lines.push(offset + "}");
  _$jscoverage['protocol/reply/Part.js'][125]++;
  return lines.join("");
});
_$jscoverage['protocol/reply/Part.js'].source = ["// Copyright 2013 SAP AG.","//","// Licensed under the Apache License, Version 2.0 (the \"License\");","// you may not use this file except in compliance with the License.","// You may obtain a copy of the License at","//","// http: //www.apache.org/licenses/LICENSE-2.0","//","// Unless required by applicable law or agreed to in writing,","// software distributed under the License is distributed on an","// \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,","// either express or implied. See the License for the specific","// language governing permissions and limitations under the License.","'use strict';","","var util = require('../../util');","var common = require('../common');","var PartKind = common.PartKind;","","var PART_HEADER_LENGTH = common.PART_HEADER_LENGTH;","var BIG_ARGUMENT_COUNT_INDICATOR = -1;","","module.exports = Part;","","function Part(kind, attributes, argumentCount, buffer, encoding) {","  this.kind = kind || PartKind.NIL;","  this.attributes = attributes || 0;","  this.argumentCount = argumentCount || 0;","  if (util.isString(buffer)) {","    this.buffer = new Buffer(buffer, encoding || 'hex');","  } else {","    this.buffer = buffer;","  }","}","","Object.defineProperty(Part.prototype, 'byteLength', {","  get: function getByteLength() {","    var byteLength = PART_HEADER_LENGTH;","    if (Buffer.isBuffer(this.buffer)) {","      byteLength += util.alignLength(this.buffer.length, 8);","    }","    return byteLength;","  }","});","","Part.create = createPart;","Part.read = readPart;","","function createPart(buffer, offset) {","  var part = new Part();","  readPart.call(part, buffer, offset);","  return part;","}","","function readPart(buffer, offset) {","  /* jshint validthis:true */","  offset = offset || 0;","","  this.kind = buffer[offset];","  this.attributes = buffer[offset + 1];","  this.argumentCount = buffer.readInt16LE(offset + 2);","  if (this.argumentCount === BIG_ARGUMENT_COUNT_INDICATOR) {","    this.argumentCount = buffer.readInt32LE(offset + 4);","  }","  var length = buffer.readInt32LE(offset + 8);","  offset += PART_HEADER_LENGTH;","  if (length &gt; 0) {","    this.buffer = new Buffer(length);","    buffer.copy(this.buffer, 0, offset, offset + length);","    offset += util.alignLength(length, 8);","  }","  return offset;","}","","Part.prototype.toBuffer = function toBuffer(size) {","  var byteLength = util.alignLength(this.buffer.length, 8);","  var buffer = new Buffer(PART_HEADER_LENGTH + byteLength);","  // Part kind, specifies nature of part data","  buffer.writeInt8(this.kind, 0);","  // Further attributes of part","  buffer.writeInt8(this.attributes, 1);","  // Argument count, number of elements in part data.","  buffer.writeInt16LE(this.argumentCount, 2);","  // Argument count, number of elements in part data (only for some part kinds).","  buffer.writeInt32LE(0, 4);","  // Length of part buffer in bytes","  buffer.writeInt32LE(this.buffer.length, 8);","  // Length in packet remaining without this part.","  buffer.writeInt32LE(size, 12);","  this.buffer.copy(buffer, PART_HEADER_LENGTH);","  if (this.buffer.length &lt; byteLength) {","    buffer.fill(0x00, PART_HEADER_LENGTH + this.buffer.length);","  }","  return buffer;","};","","Part.prototype.inspect = function inspect(options) {","  var lines = [];","  options = options || {};","  var offset = new Buffer(options.indentOffset || 0);","  offset.fill(0x20);","  offset = offset.toString('ascii');","  var kindName = common.PartKindName[this.kind];","  lines.push(offset + '{\\n');","  lines.push(offset + '  kind: PartKind.' + kindName + ',\\n');","  lines.push(offset + '  argumentCount: ' + this.argumentCount + ',\\n');","  lines.push(offset + '  attributes: ' + this.attributes + ',\\n');","  if (Buffer.isBuffer(this.buffer)) {","    var length = this.buffer.length;","    var start, end, chunk;","    start = 0;","    var hexstr = [];","    while (start &lt; length) {","      end = start + 32 &gt; length ? length : start + 32;","      chunk = this.buffer.toString('hex', start, end);","      hexstr.push(offset + '    \\'' + chunk + '\\'');","      start = end;","    }","    lines.push(offset + '  buffer: new Buffer(\\n');","    lines.push(hexstr.join(' +\\n') + ', \\'hex\\')\\n');","  } else {","    lines.push(offset + '  buffer: null\\n');","  }","  lines.push(offset + '}');","  return lines.join('');","};"];
