// syracuse-batch/lib/representations/batchServer.js
"use strict";
exports.representation = {
	resources: function() {
		var locale = require('streamline-locale');
		var globals = require('streamline-runtime').globals;
		return locale.resources(module, globals.context && globals.context.sessionLocale)();
	},
	$entityName: "batchServer",
	$facets: {
		$details: {
			$layout: {
				$layoutType: "row",
				$items: [{
					$items: [{
						$category: "section",
						$title: "{@InfoSectionTitle}",
						$layout: {
							$items: [{
								$bind: "code"
							}, {
								$bind: "auto"
							}, {
								$bind: "status"
							}]
						}
					}, {
						$category: "section",
						$title: "{@ContextSectionTitle}",
						$layout: {
							$items: [{
								$bind: "x3solution"
							}, {
								$bind: "refEndpoint"
							}, {
								$bind: "user"
							}, {
								$bind: "role"
							}, {
								$bind: "localePref"
							}]
						}
					}]
				}, {
					$items: [{
						$category: "section",
						$title: "{@ConfigSectionTitle}",
						$layout: {
							$items: [{
								$bind: "execInterval"
							}, {
								$bind: "killOverTime"
							}, {
								$bind: "maxDelay"
							}]
						}
					}, {
						$category: "section",
						$title: "{@ParametersSectionTitle}",
						$layout: {
							$items: [{
								$bind: "runtimes"
							}]
						}
					}]
				}]

			}
		},
		$edit: {
			$copy: "$details"
		},
		$query: {
			$layout: {
				$items: [{
					$category: "section",
					$layout: {
						$items: [{
							$category: "section",
							$layout: {
								$items: [{
									$bind: "$resources",
									$layout: {
										$items: [{
											$bind: "code",
										}, {
											$bind: "auto",
										}, {
											$bind: "status"
										}, {
											$bind: "x3solution"
										}, {
											$bind: "refEndpoint"
										}, {
											$bind: "user"
										}, {
											$bind: "role"
										}, {
											$bind: "localePref"
										}, {
											$bind: "execInterval"
										}, {
											$bind: "killOverTime"
										}, {
											$bind: "maxDelay"
										}]
									}
								}]
							}
						}]
					}
				}]
			}
		},
		$lookup: {
			$copy: "$query"
		},
		$select: {
			$copy: "$query"
		}
	}
};
