// syracuse-ui/lib/site/help/legal/legalPage.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var DesktopPage = require("syracuse-ui/lib/page/desktopPage").DesktopPage;
var _$eulaUrl = "/syracuse-ui/lib/site/help/legal/SAGE X3 EULA.pdf";

function LegalPage() {}

exports.LegalPage = helpers.defineClass(LegalPage, DesktopPage, {
	loadBox: function($prototype) {
		this.isMenuBarDisabled = true;
		this.isFooterDisabled = true;
		this.layoutSlot = document.createElement("div");
		this.$prototype = $prototype;
		this.$prototype.$properties.$copyright = {
			$type: "application/x-string",
			$isReadOnly: true
		};

		this.$prototype.$properties.$banner = {
			$type: "application/x-string"
		};


		this.$prototype.$links = this.$prototype.$links || {};
		this.$prototype.$links.$eula = {
			$title: syra_local.legalEula,
			$target: "blank",
			$url: _$eulaUrl
		};
		this.$prototype.$links.$banner = {
			$isDisabled: true,
			$title: "Legal banner"
		};

		var $skin = "s-site-legal";

		this.$item = {
			$category: "page",
			$css: $skin + "-header",
			$title: syra_local.legalPageTitle,
			$layout: {
				$items: [{
					$inplace: true,
					$css: $skin + "-copyright-item",
					$bind: "$copyright"
				}, {
					$format: "cards",
					$clientId: "s-legal-page-list",
					$addMaximize: false,
					$isTitleHidden: true,
					$isPagerHidden: true,
					$isQuickDesignerDisabled: true,
					$isRowIndexVisible: false,
					$selectByRowIndex: false,
					$isMenuRecordHidden: true,
					$bind: "$resources",
					$isHoverDisabled: true,
					$skin: $skin + "-card",
					renderRecordContent: function(record) {
						var head = document.createElement("div");
						head.className = "s-legal-item-head";
						var div = head.appendChild(document.createElement("div"));
						div.className = "s-legal-item-product";
						div.textContent = record.dataset.product;
						div = head.appendChild(document.createElement("div"));
						div.className = "s-legal-item-version";
						div.textContent = record.dataset.productVersion;
						record.body.appendChild(head);
						var $rows = [
							["serial", "valEnd"],
							["licenseeName", "resellerName"],
							["licenseeRegistrationNumber", "resellerRegistrationNumber"]
						];
						var tb = document.createElement("div");
						tb.className = "s-legal-item-body";
						for (var ii = 0, jj = $rows.length; ii < jj; ii++) {
							var $cells = $rows[ii];
							var row = document.createElement("div");
							row.className = "s-legal-item-row";
							for (var mm = 0, kk = $cells.length; mm < kk; mm++) {
								if (mm > 0) {
									row.appendChild(document.createElement("div")).className = "s-legal-item-field-sep";
								}
								var $bind = $cells[mm];
								var cell = document.createElement("div");
								cell.className = "s-legal-item-field-title";
								cell.textContent = record.$prototype.$properties[$bind].$title;
								row.appendChild(cell);
								cell = document.createElement("div");
								cell.className = "s-legal-item-field-value";
								record.page.loadNewItem(cell, {
									$bind: $bind,
									$inplace: true
								}, record);
								row.appendChild(cell);
							}
							tb.appendChild(row);
						}
						record.body.appendChild(tb);
					},
					$layout: {
						$items: []
					}

				}, {
					$category: "link",
					$skin: "s-nav-menu-link",
					$bind: "$eula"
				}]
			}
		};
		DesktopPage.prototype.loadBox.call(this);
		var bannerSlot = document.createElement("div");
		bannerSlot.className = "s-site-legal-banner-slot";
		var bannerItem = document.createElement("div");
		bannerItem.className = "s-site-legal-banner-item";
		bannerSlot.appendChild(bannerItem);
		this.layoutContent.domItem.insertBefore(bannerSlot, this.layoutContent.domItem.firstChild);
		var list = this.idMap["s-legal-page-list"];
		if (list.topbar) {
			list.topbar.style.display = "none";
		}
	},
	applyChange: function(newData) {
		if (this.dataset && !this.dataset.$copyright) {
			newData.$copyright = syra_local.legalCopyright;
		}
		if (newData && newData.$resources && newData.$resources.length > 0) {
			for (var ii = 0, jj = newData.$resources.length; ii < jj; ii++) {
				newData.$resources[ii].productVersion = syra_local.legalProductVersionLabel + " " + newData.$resources[ii].productVersion;
			}
		}
		DesktopPage.prototype.applyChange.call(this, newData);
	},
	onMenuClick: function(menuItem) {
		if (menuItem.$bind == "$eula") {
			window.open(_$eulaUrl, "_blank");
			return false;
		}
		return true;
	}
});
