// syracuse-ui/lib/site/help/legal/legalPage.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var DesktopPage = require("syracuse-ui/lib/page/desktopPage").DesktopPage;
var $eulaUrl = "/syracuse-ui/lib/site/help/legal/SAGE X3 EULA.pdf";
var $eulaDepsUrl = "/syracuse-main/html/main.html?url=/sdata/syracuse/collaboration/syracuse/copyrights?representation=copyright.$query&count&orderBy=name ASC";

function LegalPage() {}

exports.LegalPage = helpers.defineClass(LegalPage, DesktopPage, {
	loadBox: function($prototype) {
		this.isMenuBarDisabled = true;
		this.isFooterDisabled = true;
		this.layoutSlot = document.createElement("div");
		this.$prototype = $prototype;
		this.$prototype.$properties.$copyright = {
			$type: "application/x-string",
			$isReadOnly: true
		};

		this.$prototype.$links = this.$prototype.$links || {};
		this.$prototype.$links.$eula = {
			$title: syra_local.legalEula,
			$url: ""
		};
		this.$prototype.$links.$eulaDeps = {
			$title: syra_local.legalDepsTitle,
			$description: syra_local.legalDepsDescription,
			$url: ""
		};

		var $skin = "s-site-legal";

		this.$item = {
			$category: "page",
			$css: $skin + "-header",
			$title: syra_local.legalPageTitle,
			$layout: {
				$css: $skin + "-body",
				$layout: {
					$items: [{
						$format: "cards",
						$isTitleHidden: true,
						$isPagerHidden: true,
						$isQuickDesignerDisabled: true,
						$isRowIndexVisible: false,
						$selectByRowIndex: false,
						$isMenuRecordHidden: true,
						$bind: "$resources",
						$isHoverDisabled: true,
						$skin: $skin + "-card",
						$layout: {
							$isTitleHidden: true,
							$items: [{
								$layoutType: "stack",
								$items: [{
									$isTitleHidden: true,
									$skin: $skin + "-version",
									$layout: {
										$layoutType: "stack",
										$isFixedWidth: true,
										$widths: "*,*",
										$items: [{
											$bind: "product",
											$isTitleHidden: true,
											$isMenusHidden: true,
											$skin: $skin + "-version-item"
										}, {
											$bind: "productVersion",
											$skin: $skin + "-update-item",
											$isTitleHidden: true
										}]
									}
								}, {
									$isTitleHidden: true,
									$layout: {
										$layoutType: "row",
										$items: [{
											$isTitleHidden: true,
											$layout: {
												$layoutType: "stack",
												$items: [{
													$css: $skin + "-license-data",
													$bind: "serial"
												}, {
													$css: $skin + "-license-data",
													$bind: "licenseeName"
												}, {
													$css: $skin + "-license-data",
													$bind: "licenseeRegistrationNumber"
												}]
											}
										}, {
											$isTitleHidden: true,
											$layout: {
												$layoutType: "stack",
												$items: [{
													$css: $skin + "-license-data",
													$bind: "valEnd"
												}, {
													$css: $skin + "-license-data",
													$bind: "resellerName"
												}, {
													$css: $skin + "-license-data",
													$bind: "resellerRegistrationNumber"
												}]
											}
										}]
									}
								}]
							}]
						}
					}]
				}

			}
		};
		DesktopPage.prototype.loadBox.call(this);
		syra_dom.display(this.header, false);
		var footer = syra_dom.addDiv("s-legal-footer", this.layoutContent.domItem);
		var warning = "Warning: This software is protected by the Intellectual Property Code, copyright laws and international treaties.";
		warning += " Any improper use of the software, including any partial or total reproduction or distribution of the software or any use beyond the rights acquired on the software is strictly prohibited.";
		warning += " Anyone not complying with these provisions shall be guilty of copyright infringement and will be liable to criminal penalties provided by the law.";
		syra_dom.addTextInfo("s-site-legal-warning", warning, footer);

		this.loadNewItem(footer, {
			$isTitleHidden: true,
			$skin: "s-site-legal-copyright-item",
			$bind: "$copyright"
		});

		this.loadNewItem(footer, {
			$category: "link",
			$skin: "s-site-legal-copyright",
			$bind: "$eula"
		});
		this.loadNewItem(footer, {
			$category: "link",
			$skin: "s-site-legal-copyright",
			$bind: "$eulaDeps"
		});


	},
	applyChange: function(newData) {
		// append copyright data
		if (this.dataset && !this.dataset.$copyright) {
			newData.$copyright = syra_local.legalCopyright;
		}

		if (newData && newData.$resources && newData.$resources.length > 0) {
			for (var ii = 0, jj = newData.$resources.length; ii < jj; ii++) {
				newData.$resources[ii].productVersion = syra_local.legalProductUpdateLabel + " " + newData.$resources[ii].productVersion;
			}
		}

		DesktopPage.prototype.applyChange.call(this, newData);
	},
	onMenuClick: function(menuItem) {
		if (menuItem.$bind == "$eula") {
			window.open($eulaUrl, "_blank");
			return false;
		}
		if (menuItem.$bind == "$eulaDeps") {
			window.open($eulaDepsUrl, "_blank");
			return false;
		}
		return true;
	}

});
