// syracuse-tablet/html/js/main.js
"use strict";

//Initialization Sage Mobile jQuery plugins and native prototype object  (add method)
require('syracuse-tablet/html/js/helpers/nativePrototypes');
require('syracuse-tablet/html/js/helpers/jqueryPlugins');
var modal = require('syracuse-tablet/html/js/ui/modal');
var templates = require('syracuse-tablet/html/js/helpers/templating');
var log = require('syracuse-tablet/html/js/helpers/logger').log;
var utils = require('syracuse-tablet/html/js/helpers/utils');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var factory = require('syracuse-tablet/html/js/factories/appFactory');
var user = require('syracuse-tablet/html/js/application/user');
var auth = require('syracuse-tablet/html/js/application/authentication');


window.onerror = function(errorMsg, url, lineNumber, error) {
	try {
		modal.error("Javascript error", {
			where: "window.onerror",
			exception: error,
			message: errorMsg + "\n" + (url ? "Url : " + utils.cleanStack(url) + "\nLineNumber : " + lineNumber : "")
		});
	} catch (e) {
		var msg = errorMsg + "\n" + (error ? +JSON.stringify(error, null, 2) : (url ? "Url : " + url + "\nLineNumber : " + lineNumber : ""));
		alert(msg);
	}
};

var _init = function(lang, rsrc) {
	try {
		log && log("Init application", "history.length=" + history.length);
		// Remove templates on reload otherwise they not refreshed
		templates.removeTmpl();
		modal.init();
		locale.init(lang, rsrc);
		auth.check().always(function(userLoggedIn) {
			// Set Global smApp
			var app = window.smApp = factory.createApplication($("#application"));
			// Goto page
			var page = window.location.hash,
				options = {};
			if (userLoggedIn !== true) {
				page = "login";
				options.context = {
					redirect: window.location.hash,
					reason: "401"
				};
			}
			app.changePage(page, options).then(function(dstPage) {
				window.history.replaceState(dstPage.state, dstPage.state.uuid, "");
				log && log("Load after change page", "history.length=" + history.length);
				app.$elmt.show();
			});
		});
	} catch (e) {
		modal.error(locale.text('err.load.main'), utils.toDiagnose({
			where: "$(document).ready",
			exception: e
		}));
	}
};

exports.main = function(lang, rsrc) {
	$(document).ready(function() {
		_init(lang, rsrc);
	});
};
