// syracuse-tablet/html/js/main.js
"use strict";

//Initialization Sage Mobile jQuery plugins and native prototype object  (add method)
var modal = require('syracuse-tablet/html/js/ui/modal');
var uiutils = require('syracuse-tablet/html/js/ui/uiUtils');
var templates = require('syracuse-tablet/html/js/helpers/templating');
var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("main");
var globals = require('syracuse-tablet/html/js/helpers/globals');
var utils = require('syracuse-tablet/html/js/helpers/utils');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var appFactoryDeps = require('syracuse-tablet/html/js/application/appFactoryDeps');
var factory = require('syracuse-tablet/html/js/application/appFactory');
var auth = require('syracuse-tablet/html/js/application/authentication');
var dispatcher = require('syracuse-tablet/html/js/sdata/sdataDispatcher');
var notifications = require('syracuse-tablet/html/js/helpers/notifications');
var pageHelper = require('syracuse-tablet/html/js/application/pageHelper');

window.onerror = function(errorMsg, url, lineNumber, error) {
	uiutils.waitStop();
	try {
		modal.error("Javascript error", {
			where: "window.onerror",
			exception: error,
			message: errorMsg + "\n" + (url ? "Url : " + utils.cleanStack(url) + "\nLineNumber : " + lineNumber : "")
		});
	} catch (e) {
		var msg = errorMsg + "\n" + (error ? +JSON.stringify(error, null, 2) : (url ? "Url : " + url + "\nLineNumber : " + lineNumber : ""));
		alert(msg);
	}
};

var _initModules = function(lang, rsrc) {
	/* We need resources when we start application */
	locale.init(lang, rsrc);
	/* Init globals module - dvlpMode = true */
	globals.init(true);
	/* Init modules that are initialized by factories */
	appFactoryDeps.init();
	/* templating initialization*/
	templates.init();

	pageHelper.registerAppPages();
	/* modal initialization - load modal templates*/
	modal.init();
};
var _init = function(lang, rsrc) {
	var app;
	var fail = function(e, msg, callBack) {
		uiutils.waitStop();
		modal.error(locale.text('err.load.main'), utils.toDiagnose({
			where: "$(document).ready",
			message: msg,
			exception: e
		}), callBack);
	};
	var gotoLogin = function(reason) {
		if (app) {
			app.changePage("login", {
				reason: reason
			});
		}
	};
	try {
		uiutils.waitStart();
		log && log("Init application", "history.length=" + window.history.length);
		_initModules(lang, rsrc);
		auth.check().then(function(userLoggedIn) {
			/* Check succeeded */
			app = factory.createApplication($("#application"));
			globals.setApplication(app);
			if (userLoggedIn) {
				log && log("user logged in");
				dispatcher.getUserProfile().then(function(profile) {
					/* Profile succeeded */
					notifications.publish("sm.login", profile);
					uiutils.waitStop();
				}, function(e) {
					/* Profile failed */
					fail(e, "Get user profile failed", function() {
						auth.logout().then(function() {
							gotoLogin("Can't read user profile, please try login again");
						}, function(e) {
							fail(e, "Logout failed", function() {
								gotoLogin("Can't read user profile, please try login again");
							});
						});
					});
				});
			} else {
				log && log("user not logged in");
				gotoLogin("Please enter your credentials");
			}
		}, function(e) {
			/* Check rejected */
			fail(e, "Check login failed");
		});
	} catch (e) {
		fail(e, "Javascript exception");
	}
};

exports.main = function(lang, rsrc) {
	$(document).ready(function() {
		if (document.location.href.indexOf("?circular") > -1) {
			xrayquire.showCycles(undefined, true);
		} else {
			_init(lang, rsrc);
		}
	});
};
