// syracuse-tablet/html/js/main.js
"use strict";

//Initialization Sage Mobile jQuery plugins and native prototype object  (add method)
require('syracuse-tablet/html/js/helpers/nativePrototypes');
require('syracuse-tablet/html/js/helpers/jqueryPlugins');
var modal = require('syracuse-tablet/html/js/ui/modal');
var templates = require('syracuse-tablet/html/js/helpers/templating');
var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("main");
var globals = require('syracuse-tablet/html/js/helpers/globals');
var utils = require('syracuse-tablet/html/js/helpers/utils');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var factory = require('syracuse-tablet/html/js/factories/appFactory');
var auth = require('syracuse-tablet/html/js/application/authentication');
var dispatcher = require('syracuse-tablet/html/js/sdata/sdataDispatcher');
var notifications = require('syracuse-tablet/html/js/helpers/notifications');

window.onerror = function(errorMsg, url, lineNumber, error) {
	try {
		modal.error("Javascript error", {
			where: "window.onerror",
			exception: error,
			message: errorMsg + "\n" + (url ? "Url : " + utils.cleanStack(url) + "\nLineNumber : " + lineNumber : "")
		});
	} catch (e) {
		var msg = errorMsg + "\n" + (error ? +JSON.stringify(error, null, 2) : (url ? "Url : " + url + "\nLineNumber : " + lineNumber : ""));
		alert(msg);
	}
};

var _init = function(lang, rsrc) {
	try {
		log && log("Init application", "history.length=" + history.length);
		locale.init(lang, rsrc);
		// Remove templates on reload otherwise they not refreshed
		if (globals.dvlpMode()) {
			// Remove templates to force reload in development mode
			templates.remove();
		}
		modal.init();
		auth.check().always(function(userLoggedIn) {
			var app = factory.createApplication($("#application"));
			globals.application(app);
			if (userLoggedIn) {
				log && log("user logged in");
				dispatcher.getUserProfile().always(function(profile) {
					if (profile) {
						notifications.publish("sm.login", profile);
					} else {
						auth.logout().always(function() {
							app.changePage("login", {
								reason: "Can't read user profile, please try login again"
							});
						});
					}
				});
			} else {
				log && log("user not logged in");
				app.changePage("login", {
					reason: "Please enter your credentials"
				});
			}
		});
	} catch (e) {
		modal.error(locale.text('err.load.main'), utils.toDiagnose({
			where: "$(document).ready",
			exception: e
		}));
	}
};

exports.main = function(lang, rsrc) {
	$(document).ready(function() {
		_init(lang, rsrc);
	});
};
