// @sage/syracuse-react/src/profile/panel/profilePanel.tsx
"use strict";
import * as React from 'react';
import { PopupLink } from '../../over/popup/popupLink';
import { getLocalString } from '../../culture/localString';
import { Link } from '../../link/link';
import { UserTab } from './tabs/userTab';
import { LocalesTabController, ILocalesTabProps } from './tabs/localesTabController';
import { ThemesTab } from './tabs/themesTab';
import { IProfilePanelProps, IUserTabProps } from '../bar/profileBarProps';
import './profilePanel.less';


export class ProfilePanelLink extends React.Component<IProfilePanelProps, {}>{
	onPopupRender = () => {
		return <ProfilePanel {...this.props} />;
	}
	render() {
		return <PopupLink
			data-s-profile-panel="1"
			title={this.props.profile.fullName}
			className="s_profile_bar_textlink s_profile_bar_textlink_profile_panel"
			popup={{
				my: "right top",
				at: "right bottom",
				closeBy: ["outClick"],
				render: this.onPopupRender
			}} />;
	}
}

class ProfilePanel extends React.Component<IProfilePanelProps, {
	selectedTab: string
}>
{
	constructor(props: any, context: any) {
		super(props, context);
		this.state = {
			selectedTab: 'USER'
		};
	}
	handleClick = (value: any) => {
		this.setState({
			selectedTab: value
		});
	}
	renderSelectedTab() {
		switch (this.state.selectedTab) {
			case "USER":
				return <UserTab {...this.props as IUserTabProps} />;
			case "LOCALE":
				return <LocalesTabController {...this.props as ILocalesTabProps} />;
			case "THEME":
				return <ThemesTab selected={this.props.profile.selectedTheme}
					themes={this.props.profile.themes}
					onSelectTheme={this.props.onSelectTheme} />;
		}
	}
	renderTabLink(value: string, icon: string, title: string) {
		return <li>
			<Link
				data-s-profile-tab={value}
				className={(this.state.selectedTab === value ? "s_profile_panel_tab_selected" : "") + " s_profile_panel_tab"}
				value={value}
				title={title}
				icon={icon}
				iconText="left"
				onClick={this.handleClick}
			/>
		</li>;
	}
	render() {
		return <div className="s_profile_panel_container">
			<div className="s_profile_panel">
				<ul className="s_profile_panel_tabs">
					{this.renderTabLink("USER", "user", getLocalString("profile_user_tab"))}
					{this.renderTabLink("LOCALE", "flag", this.props.profile.selectedLocale.description)}
					{this.renderTabLink("THEME", "group", getLocalString("profile_theme_tab")
						+ " ("
						+ ((this.props.profile.selectedTheme && this.props.profile.selectedTheme.description) || getLocalString("profile_theme_default"))
						+ " )")}
					<li>
						<div className="s_profile_panel_logout_wrapper">
							<Link
								data-s-profile-logout="1"
								className="s_profile_panel_logout"
								icon="exit"
								title={getLocalString("profile_logout")}
								onClick={this.props.onLogout} />
						</div>
					</li>
				</ul>
				{this.renderSelectedTab()}
			</div>
		</div>;
	}
}
