// sprint1/lib/settings/representations/companyMob.js
exports.representation = {
    $entityName: "company",
    $facets: {
        $query: {
            $layout: {
                $items: [{
                    $category: "section",
                    $layout: {
                        $items: [{
                            $bind: "$resources",
                            $format: "cardList",
                            $cardview: {
                                $layout: {
									$layoutType: "stack",
                                    $items: [
									{
                                        $bind: "code",
                                        $isTitleHidden: true
                                    }, 
									{
                                        $bind: "description",
                                        $isTitleHidden: true
                                    }
								]
                               }
                            }
                        }]
                    }
                }]
            }
        }
    }
}
