// sprint1/lib/settings/representations/countryMob.js
exports.representation = {
    $entityName: "country",
    $facets: {
        $query: {
            $layout: {
                $items: [{
                    $category: "section",
                    $layout: {
                        $items: [{
                            $bind: "$resources",
                            $format: "cards",
                            $cards: {
                                $layout: {
                                    $layoutType: "stack",
                                    $items: [{
                                        $bind: "code",
                                        $isTitleHidden: true
                                    }, {
                                        $bind: "description",
                                        $isTitleHidden: true
                                    }]
                                }
                            }
                        }]
                    }
                }]
            }
        }
    }
}
