// sprint1/lib/settings/representations/employeeMob.js
exports.representation = {
    $entityName: "employee",
    $facets: {
        $details: {
            $layout: {
                $items: [{
                    $category: "section",
                    $layout: {
                        $items: [{
                            $title: "General",
                            $category: "section",
                            $layout: {
                                $items: [{
                                    $bind: "photo",
                                    $isTitleHidden: true
                                }]
                            }
                        }, {
                            $title: "Personal information",
                            $category: "section",
                            $layout: {
                                $items: [{
                                    $bind: "firstName"
                                }, {
                                    $bind: "lastName"
                                }, {
                                    $bind: "title"
                                }, {
                                    $bind: "maritalStatus"
                                }]
                            }
                        }, {
                            $title: "Children",
                            $category: "section",
                            $layout: {
                                $items: [{
                                    $bind: "children",
                                    $isTitleHidden: true
                                }]
                            }
                        }]
                    }
                }, {
                    $title: "Professional info",
                    $category: "section",
                    $layout: {
                        $items: [{
                            $title: "Address",
                            $category: "section",
                            $layout: {
                                $items: [{
                                    $bind: "addressStreet"
                                }, {
                                    $bind: "addressZipCode"
                                }, {
                                    $bind: "addressCity"
                                }, {
                                    $bind: "country"
                                }]
                            }
                        }, {
                            $title: "Misc.",
                            $category: "section",
                            $layout: {
                                $items: [{
                                    $bind: "foreignerWorker"
                                }, {
                                    $bind: "email"
                                }, {
                                    $bind: "phoneNumber"
                                }, {
                                    $bind: "company"
                                }]
                            }
                        }]
                    }
                }]
            }
        },
        $edit: {
            $copy: "$details"
        },
        $query: {
            $layout: {
                $items: [{
                    $category: "section",
                    $layout: {
                        $items: [{
                            $bind: "$resources",
                            $format: "cards",
                            $isPagerHidden: true,
							$layout: {
								$items: [{
									$category: "section",
									$layout: {
										$layoutType: "side",
										$items: [{
											$category: "section",
											$layoutType: "stack",
											$items: [{
												$bind: "photo",
												$isTitleHidden: true
											}]
										}, {
											$layoutType: "stack",
											$items: [{
												$expression: "{firstName} <strong>{lastName}</strong>"
											}, {
												$bind: "company",
												$isTitleHidden: true
											}]
										}]
									}
								}]
							}
                        }]
                    }
                }]
            }
        }
    }
};
