// sprint1/lib/settings/representations/employee.js
exports.representation = {
    $entityName: "employee",
    $facets: {
        $details: {
            $layout: {
                $items: [{
                    $title: "General",
                    $category: "section",
                    $layout: {
                        $items: [{
                            $title: "General",
                            $category: "block",
                            $layout: {
                                $items: [{
                                    $bind: "photo"
                                }]
                            }
                        }, {
                            $title: "Personnal informations",
                            $category: "block",
                            $layout: {
                                $items: [{
                                    $bind: "firstName"
                                }, {
                                    $bind: "lastName"
                                }, {
                                    $bind: "title"
                                }, {
                                    $bind: "maritalStatus"
                                }, {
                                    $bind: "birthDate"
                                }, {
                                    $bind: "testTime"
                                }, {
                                    $bind: "testInteger"
                                }, {
                                    $bind: "testDecimal"
                                }]
                            }
                        }, {
                            $title: "Children",
                            $category: "block",
                            $layout: {
                                $items: [{
                                    $bind: "children"
                                }]
                            }
                        }]
                    }
                }, {
                    $title: "Professional info",
                    $category: "section",
                    $layout: {
                        $items: [{
                            $title: "Address",
                            $category: "block",
                            $layout: {
                                $items: [{
                                    $bind: "addressStreet"
                                }, {
                                    $bind: "addressZipCode"
                                }, {
                                    $bind: "addressCity"
                                }, {
                                    $bind: "country"
                                }]
                            }
                        }, {
                            $title: "Misc.",
                            $category: "block",
                            $layout: {
                                $items: [{
                                    $bind: "foreignerWorker"
                                }, {
                                    $bind: "email"
                                }, {
                                    $bind: "phoneNumber"
                                }, {
                                    $bind: "company"
                                }]
                            }
                        }]
                    }
                }, {
                    $title: "Comment",
                    $category: "section",
                    $layout: {
                        $items: [{
                        	$bind: "comment",
							$textWidth: "100%",
							$textHeight: "200px" 
                        }]
                    }
                }]
            }
        },
        $edit: {
            $copy: "$details"
        },
        $query: {
            $layout: {
                $items: [{
                    $category: "section",
                    $layout: {
                        $items: [{
                            $bind: "$resources",
                            $items: [{
                                $bind: "title"
                            }, {
                                $bind: "firstName"
                            }, {
                                $bind: "lastName"
                            }, {
                                $bind: "birthDate"
                            }, {
                                $bind: "email"
                            }, {
                                $bind: "company"
                            }, {
                                $bind: "photo"
                            }]
                        }]
                    }
                }]
            }
        }
    }
}
