// syracuse-ui/lib/common/field/alphaNumField.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Field = require('./field').Field;


function PhoneField(){
}

exports.PhoneField = helpers.defineClass(PhoneField, Field, {
    getHtmlInput: function(){
        return "<input type='tel'/>";
    },
    setDataValue: function(value, record){
        this.currentValue = value;
        if (this.$isEditMode) {
            this.$$input.val(this.currentValue);
        }
        else {
            this.$$dataValue.empty().append($("<a class='s-phone-link'/>").attr("href", "callto:" + value).text(value));
        }
    }
});

function EmailField(){
}

exports.EmailField = helpers.defineClass(EmailField, Field, {
    initialize: function(){
        this.$field.$constraints = this.$field.$constraints || {};
        if (!this.$field.$constraints.$pattern) {
            this.$field.$constraints.$pattern = "^([_a-z0-9-]+(\\.[_a-z0-9-]+)*)@([_a-z0-9-]+(\\.[_a-z0-9-]+)*)\\.([a-z]{2,6})$";
            this.$field.$constraints.$patternMsg = "Invalid email";
        }
    },
    getHtmlInput: function(){
        return "<input type='email'/>";
    },
    setDataValue: function(value, record){
        this.currentValue = value;
        if (this.$isEditMode) {
            this.$$input.val(this.currentValue);
        }
        else {
            this.$$dataValue.empty().append($("<a rel='nofollow' class='s-email-link'/>").attr("href", "mailto:" + value).text(value));
        }
    }
});
