// syracuse-ui/lib/common/field/referenceField.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Field = require('syracuse-ui/lib/common/field/field').Field;

function ReferenceField(){
}

exports.ReferenceField = helpers.defineClass(ReferenceField, Field, {
    releaseMode: function(onDispose){
        if (!onDispose) {
            if (this.codeMenu) {
                this.getArticle().removeItem(this.codeMenu, true);
            }
            delete this.currentValue;
        }
        delete this.codeMenu;
        Field.prototype.releaseMode.call(this, onDispose);
    },
    getDataType: function(){
        return this.$field.$;
    },
    onInputChange: function($$input){
        var value = $$input.val();
        if (value == "") {
            document.editFieldController.notifyChange(this, null, true);
        }
        else {
            if (this.currentValue && this.currentValue.$value != value || !this.currentValue) {
                document.editFieldController.notifyChange(this, {
                    $value: value
                });
            }
        }
    },
    render: function(){
        var self = this;
        if (self.$isEditMode) {
            var css = ((self.$item.$css) ? self.$item.$css + " " : "");
            var $skinInput = css + self.$skinInput;
            (self.$$input = $("<input type='text'/>"))[0].className = $skinInput + " " + self.$skinInput + "-ref-picker";
            
            var box = document.createElement("div");
            box.className = $skinInput + "-box";
            var $$box = $(box).append(self.$$input);
            if (!self.$item.$inplace && self.$item.$isReferenceTitleVisible !== false) {
                var div = document.createElement("div");
                div.className = $skinInput + "-ref-desc";
                self.$$refTitle = $(div);
            }
            var btn = document.createElement("a");
            btn.className = css + self.$skin + "-ref-picker";
            self.$$pickerBtn = $(btn).bind("click", function(){
                document.editFieldController.clickLookUp(self);
            }).appendTo($$box);
            self.$$dataValue.append($$box).append(self.$$refTitle);
        }
        else {
            self.codeMenu = document.itemFactory.load(self.$$dataValue, {
                $bind: "$detail",
                $isDescriptionVisible: self.$item.$isReferenceTitleVisible !== false,
                $css: self.$item.$css,
                $category: "link",
                $skin: self.$skin + "-ref-link"
            }, self.boxParent);
        }
    },
    setDataValue: function(value){
        if (this.currentValue && value) {
            if (this.currentValue.$uuid && this.currentValue.$uuid == value.$uuid) {
                return;
            }
        }
        if (value) {
            var article = this.getArticle();
            this.currentValue = {
                $uuid: value.$uuid,
                $title: value.$title || ""
            };
            //            var $keyValue = this.$field.$key ? "$key" : "$value";
            //            this.currentValue.$value = value[$keyValue] || article.parseExpression(this.$field[$keyValue], value) || "";
            var $valueField = this.$field.$value ? "$value" : "$key";
            this.currentValue.$value = value[$valueField] || article.parseExpression(this.$field[$valueField], value) || "";
            var $keyField = this.$field.$key ? "$key" : "$value";
            this.currentValue.$key = value[$keyField] || article.parseExpression(this.$field[$keyField], value) || "";
        }
        else {
            this.currentValue = {
                $title: "",
                $value: "",
                $key: ""
            };
        }
        if (this.$isEditMode) {
            //this.$$input.val(this.currentValue.$value);	// old code version
            this.$$input.attr("value", this.currentValue.$value); // hack to make it work on mobile
            if (this.$$refTitle) {
                this.$$refTitle.text(this.currentValue.$title);
            }
            else {
                this.$$input.attr("title", this.currentValue.$title);
            }
        }
        else {
            if (this.$field.$links) {
                var $details = this.$field.$links.$details;
                if ($details) {
                    $details.$title = this.currentValue.$value;
                    $details.$description = this.currentValue.$title;
                    //Temp hack
                    /*                    if (value && !value[$keyValue]) {
                     value[$keyValue] = this.currentValue.$value;
                     }*/
                    if (value && !value[$keyField]) {
                        value[$keyField] = this.currentValue.$value;
                    }
                    this.codeMenu.setMenu($details, value);
                }
            }
        }
    }
});
