// syracuse-ui/lib/site/quality/qualityController.js
"use strict";
var _domQuery = require("syracuse-ui/lib/site/quality/domQuery");

var _page = {
	setNavigation: function(page, item) {
		if (item.isField && item.codeMenu && item.codeMenu.domItem) {
			var articleParent = item.articleParent;
			while (articleParent) {
				if (articleParent.isModuleRecord) {
					item.codeMenu.domItem.setAttribute("data-s-module", articleParent.$uuid);
					break;
				}
				articleParent = articleParent.articleParent;
			}
		}
	},
	setFusion: function(page) {
		page.domItem.setAttribute("data-s-fusion-page", "1");
		var blocks = page.fusionBar && page.fusionBar.blocks;
		if (blocks) {
			for (var mm = 0, kk = blocks.length; mm < kk; mm++) {
				var block = blocks[mm];
				block.domItem.setAttribute("data-s-block", block.index);
			}
		}
	}
};

function _setHidden(dom, $isHidden) {
	if ($isHidden) {
		dom.setAttribute("data-s-hidden", "1");
	} else {
		dom.removeAttribute("data-s-hidden");
	}
}

var _field = {
	setInfo: function(field) {
		if (field.domItem) {
			_setHidden(field.domItem, field.$isHidden);
			field.domItem.setAttribute("data-s-field", field.id);
			field.textArea && field.textArea.setAttribute("data-s-field", field.id);
			field.confirmInput && field.confirmInput.setAttribute("data-s-field", field.id);
			//this._setCommon(field, field._dataValue);
			//if (field.input) {
			//  field.input.setAttribute("data-s-field", field.id);
			//                this._setCommon(field, field.input);
			//}
			if (field.$isMandatory !== undefined) {
				if (field.$isMandatory) {
					field.domItem.setAttribute("data-s-mandatory", 1);
				} else {
					if (field.domItem.hasAttribute("data-s-mandatory")) {
						field.domItem.removeAttribute("data-s-mandatory");
					}
				}
			}
			if (field.domTitle) {
				field.domTitle.setAttribute("data-s-field-title", field.getTitle());
			}
			field.domItem.setAttribute("data-s-field-edit-mode", field.$isEditMode ? 1 : 0);
			field.domItem.id = field.id;
			field.domItem.setAttribute("data-s-type", field.$field.$type);
			field.domItem.setAttribute("data-s-bind", field.domItem.syra_field_bind = field.$item.$bind || "");
			field.domItem.setAttribute("data-s-title", field.domItem.syra_field_title = field.getTitle());
			field.$field.$X3Name && field.domItem.setAttribute("data-s-x3Name", field.domItem.syra_field_x3Name = field.$field.$X3Name);

			if (field.descriptionItem) {
				field.descriptionItem.setAttribute("data-s-description", field.descriptionText);
			}

			if (field.$isEditMode) {
				field.input && field.input.setAttribute("data-s-value", "0");
			} else {
				field._dataValue && field._dataValue.setAttribute("data-s-value", "1");
			}

			this._setPickers(field);
			if (field.builder) {
				this._setChoice(field);
			}
			if (field.$valueStyle) {
				var target = field.customStyleTarget || (field.input ? field.input : field._dataValue);
				target && target.setAttribute("data-s-value-style", field.$valueStyle);
			}
		}
	},
	_setPickers: function(field) {
		if (field.mnPickers) {
			var ids = Object.keys(field.mnPickers);
			for (var ii = 0, jj = ids.length; ii < jj; ii++) {
				var picker = field.mnPickers[ids[ii]];
				picker && picker.setAttribute("data-s-picker", picker.syraPickerType);
			}
		}
	},
	_setChoice: function(field) {
		if (field._choices) {
			switch (field.$item.$format) {
				case "$combo":
					for (var ii = 0, jj = field._choices.length; ii < jj; ii++) {
						var choice = field._choices[ii];
						choice.link && choice.link.setAttribute("data-s-picker", "field-choice");
					}
					break;
				case "$button":
					for (var ii = 0, jj = field._choices.length; ii < jj; ii++) {
						var btn = field._choices[ii];
						btn.setAttribute("data-s-picker", field.$item.$icon ? "choice" : "onChoiceClick");
					}
					break;
				case "$radios":
					for (var ii = 0, jj = field._choices.length; ii < jj; ii++) {
						var btn = field._choices[ii];
						btn.checkBox.input.setAttribute("data-s-field", field.id);
					}
					break;
			}

		}
	}
};

function _ensureQualityAside() {
	var aside = syra_site.qualityAside;
	if (!aside) {
		syra_site.qualityAside = aside = document.createElement("aside");
		aside.id = "s-site-quality";
		aside.style.display = "none";
		syra_site.layoutSlot.appendChild(aside);
	}
	return aside;
}

function _openTestPage(segments) {
	syra_site.loadStyleSheet("uitest.css");
	require.async("syracuse-ui/lib/tests/pages/testPage", function(err, module) {
		try {
			var $itemPage = {
				layoutSlot: document.createElement("div"),
				$category: "page",
				openerUrlSegments: segments,
				$pageCategoryClass: module.TestPage,
				$representation: {
					$prototype: {
						$testName: segments.params.name.replace("_", "/"),
						$testCategory: segments.params.category,
						$properties: {},
						$links: {},
						$actions: {}
					},
					$article: {}
				}
			};
			$itemPage.openerUrlSegments.fullUrl = document.location.href;
			syra_site.onMainPageChange($itemPage);
		} catch (error) {
			console.error(error.message + "\n" + error.stack);
		}
	});
}


function _activateController(controller) {
	controller.isActivated = true;
	_ensureQualityAside();
	controller.logRequestCount();
	controller.logLock();
	controller.logUserProfile();

	controller.page_endChange = function(page) {
		if (page.qualityIsLoaded) {
			var map = syra_store.getMap();
			var ids = Object.keys(map);
			for (var ii = 0, jj = ids.length; ii < jj; ii++) {
				var item = map[ids[ii]];
				if (item == page) {
					page.body && page.body.setAttribute("data-s-body", "1");
				}
				if (item.page == page) {
					page.isNavigationPage && _page.setNavigation(page, item);

					item.isFusionPage && _page.setFusion(item);
					if (item.isArticle && item.domItem) {
						item.domItem.setAttribute("data-s-article", item.id);
					}
					if (item.isField || item.isArrayField) {
						_field.setInfo(item);
						if (item.isArrayField) {
							if (item.builder.allColumns) {
								for (var mm = 0, kk = item.builder.allColumns.length; mm < kk; mm++) {
									var col = item.builder.allColumns[mm];
									if (col.titleCell) {
										_setHidden(col.titleCell, col.$isHidden);
										col.titleCell.setAttribute("data-s-list-col", mm);
										col.titleCell.setAttribute("data-s-list-col-title", col.titleText);
										if (col.$bind !== undefined) {
											col.titleCell.setAttribute("data-s-list-col-bind", col.$bind);
											col.titleCell.setAttribute("data-s-bind", col.$bind);
										}
										if (col.key !== undefined) {
											col.titleCell.setAttribute("data-s-list-col-key", col.key);
										}
									}
								}
							}
							if (item.$item.$format) {
								item.domItem.setAttribute("data-s-list-format", item.$item.$format);
							}

						}
					}
					if (item.popupCard) {
						item.popupCard.mnPicker && item.popupCard.mnPicker.setAttribute("data-s-picker", item.popupCard.mnPicker.syraPickerType);
					}
					if (item.isRecordArticle) {
						if (item.isFilterArticle) {
							item.domItem.setAttribute("data-s-filter-record", item.$uuid);
						} else {
							item.domItem.setAttribute("data-s-record", item.$uuid);
							item.dataRow && item.dataRow.setAttribute("data-s-record", item.$uuid);
							//item.body && item.body.setAttribute("data-s-record", item.$uuid);
							item.freezeRow && item.freezeRow.setAttribute("data-s-record", item.$uuid);
							if (item.treeNode) {
								item.treeNode.item && item.treeNode.item.setAttribute("data-s-record", item.$uuid);
							}
							item.rowCard && item.rowCard.contentSlot && item.rowCard.contentSlot.outCardSlot.setAttribute("data-s-record", item.$uuid);
							item.outCardSlot && item.outCardSlot.setAttribute("data-s-record", item.$uuid);


							item.getRecordIndex && item.domItem.setAttribute("data-s-record-index", item.getRecordIndex());
						}
					}

				}
			}
		}
	};
	controller.page_onAfterLoad = function(page) {
		page.qualityIsLoaded = true;
		this.page_endChange(page);
		if (typeof window.callPhantom === 'function') {
			window.callPhantom({
				event: "onAfterLoad",
				autoFetch: true
			});
		}
	};
	controller.logUserProfile = function() {
		var dataset = syra_site.userProfile.dataset;
		var value = (dataset.selectedLocale && dataset.selectedLocale.$value) || "";
		var aside = _ensureQualityAside();
		aside.setAttribute('data-s-local', value);
		aside.setAttribute('data-s-role', (dataset.selectedRole && dataset.selectedRole.description) || "");
		aside.setAttribute('data-s-endpoint', (dataset.selectedEndpoint && dataset.selectedEndpoint.description) || "");
	};
	controller.logRequestCount = function(count) {
		_ensureQualityAside().setAttribute("data-s-request-count", count || 0);
	};
	controller.logLock = function(value) {
		_ensureQualityAside().setAttribute("data-s-uilock", value || 0);
	};
	controller.logFocus = function(field) {
		_ensureQualityAside().setAttribute("data-s-focus", field.id);
	};
}

exports.create = function() {
	return {
		domQuery: _domQuery,
		onAfterLogon: function(onAfterLogon) {
			this.enableTestRobot = syra_site.userProfile.dataset.enableTestRobot || window.callPhantom;
			this.enableTestRobot && _activateController(this);
			onAfterLogon();
		},
		openTestPage: function(segments) {
			if (segments && segments.$url && (segments.$url.indexOf("s-uitest") >= 0)) {
				if (!this.isActivated) {
					_activateController(this);
				}
				_openTestPage(segments);
				return true;
			}
		},
		page_endChange: function(page) {},
		page_onAfterLoad: function(page) {},
		logUserProfile: function() {},
		logRequestCount: function(count) {},
		logLock: function(value) {},
		logFocus: function(field) {}
	};
};
