// @sage/syracuse-ui/lib/office/ppt/pptInterfaceDummy.js
"use strict";
var helpers = require('@sage/syracuse-core').helpers;

function PptInterfaceDummy() {

}

exports.PptInterfaceDummy = helpers.defineClass(PptInterfaceDummy, null, {
	getResourceUrl: function() {
		return this._getBaseURIParam("url");
	},
	addDataToWorksheet: function(data) {
		var jsonData = JSON.stringify(data);
		alert(jsonData);
	},
	getDocumentContent: function() {
		return "Dummy content";
	},
	getDocumentTitle: function() {},
	setDocumentTitle: function(value) {},
	getDocumentUrl: function() {},
	setDocumentUrl: function(value) {},
	refreshNextChart: function() {},
	notifySaveDocumentDone: function(value) {},
	getActionType: function() {
		return "ppt_populate_worksheet";
	},
	getAddinVersion: function() {},
	_getBaseURIParam: function(param) {
		var params = document.baseURI.split("?")[1].split("&");
		var val;
		params.forEach(function(p) {
			var idx = p.indexOf(param + "=");
			if (idx > -1) {
				val = p.substring(idx + param.length + 1);
				val = decodeURIComponent(val);
			}
		});
		return val;
	},
	expectedVersion: function(value) {}
});
