// @sage/syracuse-react/src/test/tracker/samples.ts
"use strict";
export const samples = [{
	location: "/sdata/$trackers('63023233-9f27-41bb-9749-2331a99596ea')",
	uuid: "record0",
	$state: "Job started",
	phaseDetail: "print",
	progress: 25,
	elapsedSeconds: 10,
	remainingSeconds: 50,
	startDate: 1428699906078,
	title: "Pays",
	kind: "print",		
	$links: {
		$download: {
			$title: "Download"
		}
	},
	$diagnoses: [0, 1, 2, 3, 4].map(function(ii) {
		return {
			$message: "number " + ii,
			$severity: "error"
		};
	}),
	isProcessing: false
}, {
	uuid: "record1",
	$state: "Job started",
	phaseDetail: "printing",
	progress: 50,
	elapsedSeconds: 10,
	remainingSeconds: 50,
	startDate: 1428699906078,
	title: "Compagnies",
	kind: "print",
	$diagnoses: [{
		$message: "This is a test diagnose : warning",
		$severity: "warning",
		$stackTrace: "TypeError: Object 1115 has no method 'replace' at C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:18:39 at Array.forEach (native) at change (C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:16:20) at C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:20:41 at Array.forEach (native) at change (C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:16:20) at C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:20:41 at Array.forEach (native) at change (C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:16:20) at C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:20:41 at Array.forEach (native) at change (C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:16:20) at C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:20:41 at Array.forEach (native) at change (C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:16:20) at C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:20:41 at Array.forEach (native) at change (C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:16:20) at C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:20:41 at Array.forEach (native) "
	}],
	$links: {
		$cancel: {
			$title: "Cancel"
		}
	},
	isProcessing: true
}, {
	uuid: "record2",
	$state: "Job started",
	phaseDetail: "updating index",
	progress: 50,
	elapsedSeconds: 10,
	remainingSeconds: 50,
	startDate: 1428699906078,
	title: "Index update",
	kind: "operation",
	$links: {
		$abort: {
			$title: "Abort"
		}
	},
	isProcessing: true
}, {
	"uuid": "863626cc-a95e-4e0e-8f88-b7222f31929d",
	"location": "/sdata/$trackers('863626cc-a95e-4e0e-8f88-b7222f31929d')",
	"phase": "En cours de démarrage",
	"phaseDetail": "Appliquer la mise à jour",
	"progress": 100,
	"elapsedSeconds": 1065,
	"remainingSeconds": 0,
	"pollingMillis": 1000,
	"$diagnoses": [{
		"$severity": "info",
		"$message": "Mise à jour de X3 on X3U9SQL"
	}, {
		"$severity": "info",
		"$message": "Application du patch X3 FILPATCH DIRECTORY"
	}, {
		"$severity": "info",
		"$message": "Copie du patch vers PATCH/FILPATCH"
	}, {
		"$severity": "info",
		"$message": "Nettoyage"
	}, {
		"$severity": "info",
		"$message": "Copie des fichiers vers PATCH/FILPATCH"
	}, {
		"$severity": "info",
		"$message": "mkdir PATCH"
	}, {
		"$severity": "info",
		"$message": "mkdir PATCH/FILPATCH"
	}, {
		"$severity": "warning",
		"$message": "Le patch de l'update 9.0.5 ne peut être appliqué au dossier X3 (9.0.5)"
	}, {
		"$severity": "error",
		"$message": "Le patch de l'update 9.0.5 ne peut être appliqué au dossier X3 (9.0.5)"
	}, {
		"$severity": "info",
		"$message": "Application du patch X3 à SEEDWMS"
	}, {
		"$severity": "info",
		"$message": "Application du patch X3 à SEED"
	}, {
		"$severity": "info",
		"$message": "Arrêt du serveur batch"
	}, {
		"$severity": "info",
		"$message": "Arrêt du serveur batch"
	}, {
		"$severity": "info",
		"$message": "Démarrage du serveur batch"
	}, {
		"$severity": "info",
		"$message": "Démarrage du serveur batch"
	}, {
		"$severity": "info",
		"$message": "Le serveur batch est démarré"
	}, {
		"$severity": "info",
		"$message": "Le serveur batch est démarré"
	}, {
		"$severity": "info",
		"$message": "TX_00045_190 being integrated on SEED"
	}, {
		"$severity": "info",
		"$message": "TX_00045_190 being integrated on SEEDWMS"
	}],
	"$links": {},
	"startDate": 1484583121193
}];
