// syracuse-ui/lib/article/layouts/authoring/layoutSettings.js
"use strict";
var locale = require('syracuse-core/lib/locale');

exports.$columns = ["50,50", "25,75", "75,25", "38,62", "62,38", "25,25,50", "25,50,25", "50,25,25", "33,33,33", "25,25,25,25"];
var _$composite = {
    "composite-3-1": {
        $layoutType: "stack",
        $items: [{
            $layoutType: "columns",
            $layoutSubType: "50,50",
            $items: [{
                $layoutType: "stack",
                $items: []
            }, {
                $layoutType: "stack",
                $items: []
            }]
        }, {
            $layoutType: "tabs",
            $items: []
        }, {
            $layoutType: "columns",
            $layoutSubType: "50,50",
            $items: [{
                $layoutType: "stack",
                $items: []
            }, {
                $layoutType: "stack",
                $items: []
            }]
        }]
    },
    "composite-3-2": {
        $layoutType: "stack",
        $items: [{
            $layoutType: "columns",
            $layoutSubType: "50,50",
            $items: [{
                $layoutType: "stack",
                $items: []
            }, {
                $layoutType: "stack",
                $items: []
            }]
        }, {
            $layoutType: "columns",
            $layoutSubType: "50,50",
            $items: [{
                $layoutType: "stack",
                $items: []
            }, {
                $layoutType: "stack",
                $items: []
            }]
        }, {
            $layoutType: "columns",
            $layoutSubType: "50,50",
            $items: [{
                $layoutType: "stack",
                $items: []
            }, {
                $layoutType: "stack",
                $items: []
            }]
        }]
    },
    "composite-4-1": {
        $items: [{
            $layoutType: "columns",
            $layoutSubType: "50,50",
            $items: [{
                $layoutType: "stack",
                $items: []
            }, {
                $layoutType: "stack",
                $items: []
            }]
        }, {
            $layoutType: "columns",
            $layoutSubType: "50,50",
            $items: [{
                $layoutType: "stack",
                $items: []
            }, {
                $layoutType: "stack",
                $items: []
            }]
        }, {
            $layoutType: "columns",
            $layoutSubType: "50,50",
            $items: [{
                $layoutType: "stack",
                $items: []
            }, {
                $layoutType: "stack",
                $items: []
            }]
        }, {
            $layoutType: "columns",
            $layoutSubType: "50,50",
            $items: [{
                $layoutType: "stack",
                $items: []
            }, {
                $layoutType: "stack",
                $items: []
            }]
        }]
    },
    "composite-1-3": {
        $layoutType: "columns",
        $layoutSubType: "50,50",
        $items: [{
            $layoutType: "stack",
            $items: []
        }, {
            $layoutType: "stack",
            $items: [{
                $layoutType: "columns",
                $layoutSubType: "50,50",
                $items: [{
                    $layoutType: "stack",
                    $items: []
                }, {
                    $layoutType: "stack",
                    $items: []
                }]
            }, {
                $layoutType: "columns",
                $layoutSubType: "50,50",
                $items: [{
                    $layoutType: "stack",
                    $items: []
                }, {
                    $layoutType: "stack",
                    $items: []
                }]
            }, {
                $layoutType: "columns",
                $layoutSubType: "50,50",
                $items: [{
                    $layoutType: "stack",
                    $items: []
                }, {
                    $layoutType: "stack",
                    $items: []
                }]
            }]
        }]
    }
};


exports.getCompositeTypes = function(){
    return Object.keys(_$composite);
};
exports.getComposite = function($layoutSubType){
    var $layout = _$composite[$layoutSubType];
    $layout.$title = locale.resources(module)()["ls_title_" + $layoutSubType.replace("-", "_", "g")];
    return $layout;
}
