// syracuse-collaboration/lib/contract.js
exports.contract = {
    application: "syracuse",
    contract: "collaboration",
    resources: require('./strings').resources,
    entities: {
        user: require('./entities/user').entity,
        group: require('./entities/group').entity,
        endPoint: require('./entities/endPoint').entity,
        role: require('./entities/role').entity,
        userEndpoint: require('./entities/userEndpoint').entity,
        localePreference: require('./entities/localePreference').entity,
        userProfile: require('./entities/userProfile').entity,
        userPreference: require('./entities/userPreference').entity,
        application: require('./entities/application').entity,
        x3server: require('./entities/x3server').entity,
        oauth2: require('./entities/oauth2').entity,
        ldap: require('./entities/ldap').entity,
        setting: require('./entities/setting').entity,
        // tools
        importTool: require('./entities/importTool').entity,
        searchAdmin: require('./entities/searchAdmin').entity,
        sessionInfo: require('./entities/sessionInfo').entity,
        fusionSessionInfo: require('./entities/fusionSessionInfo').entity,
		moduleVersion: require('./entities/moduleVersion').entity,
        // authoring
        configuration: require('./entities/page/configuration').entity,
        portlet: require('./entities/page/portlet').entity,
//        pagePortlet: require('./entities/page/pagePortlet').entity,
        menuItem: require('./entities/page/menuItem').entity,
        menuItemParameter: require('./entities/page/menuItemParameter').entity,
        menuCategory: require('./entities/page/menuCategory').entity,
        shell: require('./entities/page/shell').entity,
        page: require('./entities/page/page').entity,
        pageFileLoader: require('./entities/page/pageFileLoader').entity,
        pageFileLoaderItem: require('./entities/page/pageFileLoaderItem').entity,
        pageData: require('./entities/page/pageData').entity,
        pageDataHistory: require('./entities/page/pageDataHistory').entity,
        pageDataHistoryItem: require('./entities/page/pageDataHistoryItem').entity,
        dashboardDef: require('./entities/page/dashboardDef').entity,
        dashboardVariant: require('./entities/page/dashboardVariant').entity,
        dashboardVignette: require('./entities/page/dashboardVignette').entity,
        dashboardAuth: require('./entities/page/dashboardAuth').entity,
        // storage area
        team: require('./entities/team').entity,
        document: require('./entities/document').entity,
        documentTag: require('./entities/documentTag').entity,
        documentTagCategory: require('./entities/documentTagCategory').entity,
        // lookups
        lookupLocale: require('./entities/lookupLocale').entity,
        lookupX3Folder: require('./entities/lookupX3Folder').entity,
        lookupEntity: require('./entities/lookupEntity').entity,
        lookupEntityAttr: require('./entities/lookupEntityAttr').entity,
        lookupRepresentation: require('./entities/lookupRepresentation').entity,
        // excel
        excelBrowseDatasource: require('./entities/excelBrowseDatasource').entity,
        excelOrderAttr: require('./entities/excelOrderAttr').entity,
        // history
        navHistory: require('./entities/navHistory').entity,
        
        msoMailMergeDS: require('./entities/msoMailMergeDS').entity,
        msoMailMergeDSSel: require('./entities/msoMailMergeDSSel').entity,
        msoMailMergeDocSel: require('./entities/msoMailMergeDocSel').entity,
    },
    representations: {
        // representation name here, not entity name (but might be the same as the entity name)
        user: require('./representations/user').representation,
        userMob: require('./representations/userMob').representation,
        role: require('./representations/role').representation,
        roleMob: require('./representations/roleMob').representation,
        endPoint: require('./representations/endpoint').representation,
        group: require('./representations/group').representation,
        groupMob: require('./representations/groupMob').representation,
        team: require('./representations/team').representation,
        portlet: require('./representations/portlet').representation,
        menuItem: require('./representations/menuItem').representation,
        pageData: require('./representations/pageData').representation,
        dashboardDef: require('./representations/dashboardDef').representation,
        documentExcel: require('./representations/documentExcel').representation,
        excelBrowseDsEntity: require('./representations/excelBrowseDsEntity').representation,
        excelBrowseDsMenuItem: require('./representations/excelBrowseDsMenuItem').representation,
        userProfileMob: require('./representations/userProfileMob').representation,
        
        msoMailMergeDS: require('./representations/msoMailMergeDS').representation,
        msoMailMergeDSSel: require('./representations/msoMailMergeDSSel').representation,
        msoMailMergeDSSelEntity: require('./representations/msoMailMergeDSSelEntity').representation,
        msoMailMergeDSSelMenu: require('./representations/msoMailMergeDSSelMenu').representation,
        msoMailMergeDocSel: require('./representations/msoMailMergeDocSel').representation,
        msoWordDocument: require('./representations/msoWordDocument').representation,
    },
    searchFacets: {
        group: {
            $title: "Groups",
            $lookup: {
                $entity: "group",
                $keyField: "description"
            },
            $fields: {
                user: "groups",
                group: "description"
            }
        },
        role: {
            $title: "Roles",
            $lookup: {
                $entity: "role",
                $keyField: "description"
            },
            $fields: {
                group: "role",
                role: "description"
            }
        },
        menuCategory: {
            $title: "Category",
            $fields: {
                portlet: "categories",
                menuItem: "categories"
            }
        }
    },
    dbMeta: {
        version: 15,
        initScript: "syracuse-admin-init.json",
        updateScript: "syracuse-collaboration/lib/updateScript"
    }
};
