// syracuse-tablet/html/js/controlsx/ctrlBase.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var Base = require('syracuse-tablet/html/js/controlsx/structElmt').Klass;
var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("ctrl");

var _Klass = utils.defineClass(
	function(controller, article, prototype) {
		var self = this;
		self.prototype = prototype;
		Base.call(self, controller, prototype.data('$type'), article);
	}, Base, {
		defaultDisplayOptions: function() {
			var self = this;
			return {
				cssDefCtrl: ["s-m-control", "s-m-" + self.typeName + "-ctrl"]
			};
		},
		destroy: function() {
			var self = this;
			self.prototype = null;
			Base.prototype.destroy.call(self);
		},
		/**
		 * Default ctrl html  - Just the value embedded into a div
		 */
		buildHtml: function(html, controllerDao, buildOptions) {
			var self = this;
			self.buildHtmlBefore(html, controllerDao, buildOptions);
			if (controllerDao) {
				if (!controllerDao.getDisplayValue) throw new Error("BuildHtml - dao is expected for data parameter");
				html.push(controllerDao.getDisplayValue(self.article.$bind));
			}
			self.buildHtmlAfter(html, controllerDao, buildOptions);
		},
		buildHtmlBefore: function(html, controllerDao, buildOptions) {
			var self = this;
			if (self.displayOptions.link) {
				html.push('<a href="#" ');
				html.push(self.displayOptions.link);
				html.push('>');
			}
			html.push('<div class="');
			html.push(self.displayOptions.cssDefCtrl.join(' '));
			html.push('" id="');
			html.push(self.id);
			html.push('"');
			html.push('>');
		},
		buildHtmlAfter: function(html, controllerDao, buildOptions) {
			var self = this;
			html.push('</div>');
			if (self.displayOptions.link) {
				html.push('</a>');
			}
		}
	});

exports.Klass = _Klass;
