// syracuse-tablet/html/js/controls/field/ctrlQuantity.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var Base = require('syracuse-tablet/html/js/controls/field/ctrlFieldBase').Klass;
var formatApi = require('syracuse-tablet/html/js/helpers/formatApi');
var ctrlFactory = require('syracuse-tablet/html/js/controls/ctrlFactory');

var _Klass = utils.defineClass(

	/**
	 * options.isTableCell true/false
	 */
	function CtrlQuantity(controller, article, prototype, options) {
		var self = this;
		Base.call(self, controller, article, prototype, options);
		self.unitFieldInfo = prototype.getUnitFieldInfo();
		// Create extra control for UNIT in read only mode
		// !! if grid Cell we ca't create the control with createControl
		self.displayUnit = !options.isTableCell && self.unitFieldInfo && self.unitFieldInfo.unitFieldProto && !self.unitFieldInfo.unitFieldProto.data("$isExcluded");
		self.unitCtrl = null;
	}, Base, {

		initFormatter: function() {
			var valueProto = this.prototype.data('$value');
			return formatApi.getFormatter(valueProto.$type, valueProto.$format);
		},

		destroy: function() {
			var self = this;
			Base.prototype.destroy.call(self);
			if (self.unitCtrl) {
				self.unitCtrl.destroy();
				self.unitCtrl = null;
			}
		},

		buildHtml: function($$parent, controllerDao, buildOptions) {
			var self = this;
			Base.prototype.buildHtml.call(self, $$parent, controllerDao, buildOptions);
			(self.$$elmt).addClass("s-m-qty");
			if (self.displayUnit) {
				self.unitCtrl = ctrlFactory.createControl(self.controller, self, {
					"$bind": self.unitFieldInfo.unitBindName,
				}, self.unitFieldInfo.unitFieldProto, {
					isUnit: true
				});
				self.unitCtrl.buildHtml(self.$$elmt, controllerDao, buildOptions);
			}
		},

		buildFieldValue: function(fieldSlot, buildOptions) {
			var self = this;
			var valueSlot = Base.prototype.buildFieldValue.call(self, fieldSlot, buildOptions);
			var value = self.getFormattedValue(self.getValue());
			if (self.$isEditMode) {
				self.createMainInput(valueSlot, value);
			} else {
				self.appendTextValue(valueSlot, value);
			}
		}
	});

exports.Klass = _Klass;
