// syracuse-tablet/html/js/controls/field/ctrlReference.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var sdataUtils = require('syracuse-tablet/html/js/sdata/sdataUtils');
var uiUtils = require('syracuse-tablet/html/js/ui/uiUtils');
var Base = require('syracuse-tablet/html/js/controls/field/ctrlFieldBase').Klass;
var fontUtils = require('syracuse-tablet/html/js/ui/fontUtils');

/**
 *  auth example
 *  {
 *  	$value: {
 *  		$theme: "A",
 *  		$valueFirst: false,
 *  		$position: "stack",
 *  		$refDescrHidden: false,
 *  		$refDescr: {
 *  			$theme: "B",
 *  			$fontSize: "small",
 *  			$withTitle: true,
 *  			$before: true
 *  		}
 *  	}
 *  };
 */

var _Klass = utils.defineClass(

	function CtrlReference(controller, article, prototype, options) {
		var self = this;
		Base.call(self, controller, article, prototype, options);
		// See ctrlQuantity - Control that displays the unit
		self.isUnit = self.options.isUnit === true;
		if (self.isUnit) {
			self.$isEditMode = false;
			self.cssType = "s-m-unit";
		} else {
			self.cssType = "s-m-ref";
		}
	}, Base, {
		/**
		 * Init the properties to allow reuse of this control
		 */
		initReuseProperties: function() {
			Base.prototype.initReuseProperties.call(this);
			this.picker = null;
		},
		getValue: function(prop) {
			var val = Base.prototype.getValue.call(this);
			return val ? val[prop || "$value"] || "" : "";
		},
		buildFieldTitle: function(fieldSlot, buildOptions) {
			var self = this;
			return self.isUnit ? null : Base.prototype.buildFieldTitle.call(self, fieldSlot, buildOptions);
		},
		buildFieldValue: function(fieldSlot, buildOptions) {
			var self = this;
			var parentElmt = Base.prototype.buildFieldValue.call(self, fieldSlot);
			var value = self.getValue();
			if (self.$isEditMode) {
				self.buildEditValue(parentElmt, self.getValue());
			} else {
				var descr = (self.getValue("$description") || "").trim();
				var title = (self.getValue("$title") || "").trim();
				if (self.isUnit) {
					self.buildUnitValue(parentElmt, value, descr, title);
				} else {
					self.buildDetailValue(parentElmt, value, descr, title);
				}
			}
		},
		buildEditValue: function(parentElmt, value) {
			var self = this;
			var bsInputGroup = uiUtils.createDomElement("div", ["input-group"], null, null, parentElmt);
			self.createMainInput(bsInputGroup, self.getFormattedValue(value));
			var $lookuplink = self.prototype.getDataByPath("$item.$links.$lookup");
			if (!$lookuplink) {
				// build lookup link
				throw new Error("ctrlReference - Expected $lookup link");
			} else {
				// get dom $lookup link parameters/attributes
				var span = uiUtils.createDomElement("span", ["input-group-btn"], null, null, bsInputGroup);
				var attrs = self._getLinkAttrs(sdataUtils.getLinkInfo($lookuplink.$url, self.getDao()), {
					"data-param": self.id,
					"data-parent-id": self.controller.id,
					"data-control-id": self.id,
					"title": self.prototype.resolveExpression($lookuplink.$title) || "Lookup"
				});
				self.picker = uiUtils.createDomElement("a", ["btn", "btn-default", "s-m-meta"], null, attrs, span);
				uiUtils.createDomElement("span", fontUtils.refFieldIcon(), null, null, self.picker);
			}
		},
		buildDetailValue: function(parentElmt, value, descr, title) {
			var self = this;
			// build container
			var container = self.addLink(parentElmt);

			// build container of reference value if any
			if (value.length > 0) {
				// set description content according to authoring value of $refDescFormat
				// Available values are desc, title, desctitle, and titledesc. Default is "desc"
				var descContent;
				var $refDescFormat = self.getAuthoring("$refDescFormat");
				switch ($refDescFormat) {
					case "title":
						descContent = title;
						break;
					case "desctitle":
						descContent = self._combineTexts(descr, title);
						break;
					case "titledesc":
						descContent = self._combineTexts(title, descr);
						break;
						// "desc" will be considered as default
					default:
						descContent = descr;
						break;
				}

				// build content and set position according to description content authoring position $refDescPosition
				// available values are bottom, right, left, top, none. Default is "right"
				var $refDescPosition = self.getAuthoring("$refDescPosition");
				var $$value = $("<div>").addClass("s-m-ref-val").text(value);
				var $$desc = $("<div>").addClass("s-m-descr").text(descContent);
				var arr;
				switch ($refDescPosition) {
					case "bottom":
						$(container).append($$value).append($$desc);
						break;
					case "left":
						$(container).append($$desc).append($$value);
						arr = [$$value, $$desc];
						$.each(arr, function(index, $$ee) {
							$$ee.css("display", "inline-block");
						});
						$$desc.css("padding-right", "5px");
						break;
					case "top":
						$(container).append($$desc).append($$value);
						break;
					case "none":
						$(container).append($$value);
						break;
					case "right": // "right" will be considered as default
					default:
						arr = [$$value, $$desc];
						$(container).append($$value).append($$desc);
						$.each(arr, function(index, $$ee) {
							$$ee.css("display", "inline-block");
						});
						$$desc.css("padding-left", "5px");
				}
			}
		},
		buildUnitValue: function(parentElmt, value, descr, title) {
			var self = this;
			var label = descr.length > 0 ? descr : title.length > 0 ? title : value;
			// TODO See how we display unit
			label = value || "";
			if (label && label.toLowerCase) label = label.toLowerCase().smCapitalize();
			uiUtils.createDomElement("span", ["s-m-descr"], label, null, self.addLink(parentElmt));
		},
		addLink: function(parentElmt) {
			var self = this;
			if (self.options.isArrayChild === true) return parentElmt;
			var $detaillink = self.prototype.getDataByPath("$item.$links.$details");
			// if links && $links.$details, container is an anchor, otherwise it is a simple div
			if ($detaillink) {
				// get dom link parameters/attributes (we consider only $details)
				return uiUtils.createDomElement("a", null, null, self._getLinkAttrs(sdataUtils.getLinkInfo($detaillink.$url, self.getDao()), {
					"title": self.prototype.resolveExpression($detaillink.$title) || "Detail"
				}), parentElmt);
			} else {
				return parentElmt;
			}
		},
		/**
		 * Set field value after a lookup selection
		 * 	rowData: data of selected row
		 */
		setLookupValue: function(rowData) {
			var self = this;
			// Field that contains the value
			var lookupField = self.prototype.getLookupValue();
			if (lookupField) {
				var val = rowData[lookupField];
				if (val != null) {
					self.setFieldValue(val);
					$(self.domInput).focus();
				}
			}
		},
		_getLinkAttrs: function(link, options) {
			var self = this;
			var attrs = {
				"href": "#",
				"data-nav": link.page,
				"data-sdata-url": link.sDataUrl
			};
			if (self.controller.isVignette) {
				// Its the vignette which will load the page - Do not modify it - FDB
				attrs["data-nav-target"] = self.controller.openLinkInVignette() ? "vignette" : "application";
			}
			if (options) {
				for (var p in options) attrs[p] = options[p];
			}
			return attrs;
		},
		_combineTexts: function(el1, el2) {
			var res;
			// el1-el2 || el1 || el2
			res = (el1 && el2) ? (el1 === el2 ? el1 : el1 + "-" + el2) : (el1 || el2 || "");
			return res;
		}

	});

exports.Klass = _Klass;
