// syracuse-tablet/html/js/controls/field/ctrlImage.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var fontUtils = require('syracuse-tablet/html/js/ui/fontUtils');
var Base = require('syracuse-tablet/html/js/controls/field/ctrlFieldBase').Klass;
var uiUtils = require('syracuse-tablet/html/js/ui/uiUtils');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var clientContract = require('syracuse-tablet/html/js/sdata/entities/clientContract');
var native = require('syracuse-tablet/html/js/helpers/native/native');

var _handlersMap = {
	"$delete": "_onClearClick",
	"$edit": "_onEditClick",
	"$sign": "_onSignClick",
	"$camera": "_onCameraClick",
	"$select": "_onUploadClick",
};

var _templates = {
	footer: '\
			<footer> \
					{{#each actions}}\
						{{#if active}}\
							<a class="ctrl-evt-click"\
								href = "#"\
								draggable="false"\
								action="{{id}}">\
								<span \
								{{#if css}}\
									class="{{css}}"\
								{{/if}}\
								>\
								{{#if text}}\
									{{text}}\
								{{/if}}\
								</span>\
							</a>\
						{{/if}}\
					{{/each}}\
			</footer> \
		'
};
var _getHtml = function(name, ctx) {
	var key = name + "compiled";
	var tmpl = _templates[key];
	if (!tmpl) {
		tmpl = _templates[key] = Handlebars.compile(_templates[name]);
	}
	if (!tmpl) alert("template " + name + " not found");
	return tmpl(ctx);
};


/**
 * Image  field
 * authoring:
 * 		$imgHeight	set the height of the image - xsmall, small...
 */
var _Klass = utils.defineClass(

	function CtrlImage(controller, article, prototype, options) {
		Base.call(this, controller, article, prototype, options);
		this.nativePenAnnotatedImage = native.hasCapability("penAnnotatedImage");
		this.nativeSignatureImage = native.hasCapability("penSignatureImage");
		this.nativeImageCamera = native.hasCapability("camera");
		this.cssType = "s-m-img";
	}, Base, {
		initReuseProperties: function() {
			Base.prototype.initReuseProperties.call(this);
			this.$$layout = null;
		},
		applyMetaData: function(metaData, buildOptions) {
			Base.prototype.applyMetaData.call(this, metaData, buildOptions);
			if (this.$isEditMode && this.$$layout) {
				var disabled = metaData ? metaData.$isReadOnly || metaData.$isDisabled : false;
				this.$$layout.children("footer").toggle(disabled !== true);
			}
		},

		buildFieldValue: function(parentSlot, buildOptions) {
			if (!this.article.$imgHeight) {
				this.article.$imgHeight = buildOptions && (buildOptions.displayCtx === "table" || buildOptions.displayCtx === "card") ? "small" : "medium";
			}
			var value = this.getFormattedValue(this.getValue());
			// Create the image stuff under the standard s-m-value parent
			var imgParent = Base.prototype.buildFieldValue.call(this, parentSlot, buildOptions);
			var docUrl = this._getLocalUrlValue();
			if (!docUrl) {
				docUrl = this.prototype.data("$url", this.getDao()) || "";
				if (clientContract.isLocalEntityUrl(docUrl)) {
					// Used to returned data in client entities
					docUrl = clientContract.callLocalEntityUrl(docUrl);
				} else {
					//TODO use $etag or else - No need to force reload if updateLayout - data are in the browser's cache
					if (buildOptions.updateLayout !== true) {
						docUrl += '?salt=' + ((new Date()).getTime());
					}
				}
			}
			var empty = value == null || value === "" || docUrl.length === 0;
			if (empty) $(imgParent).addClass("s-m-empty");
			// if FileReader not supported by the browser, we display detail view (fix for Safari)
			if (this.$isEditMode && typeof FileReader !== "undefined") {
				this._buildEditImageField(imgParent, docUrl, empty);
			} else {
				if (this.article.$imgDisplayIcon === true) {
					// We want to display images as icons - enabled only for arrays in table mode
					var css;
					if (empty) {
						css = fontUtils.getColImageDefIcon(true).css;
					} else {
						css = (this.article.$imgIcon || fontUtils.getColImageDefIcon(false)).css;
					}
					return $(imgParent).addClass(css);
				} else if (empty) {
					// Only icon
					if (buildOptions && buildOptions.displayCtx === "card") {
						// same height for all cards -> We set the height
						this._defaultImg(imgParent, null);
					} else {
						$(imgParent).addClass(this._getDefaultIcon());
					}
				} else {
					// link - Maybe we could manage a modal to display the image 
					var e = uiUtils.createDomElement("a", ["img-responsive ctrl-evt-click"], null, {
						"href": "#"
					}, imgParent);
					// Image with link
					e = this._newImg(e, docUrl);
				}
			}
		},
		_isBtnActive: function(btnId) {
			var isSignature = this.article.$isSignature ? true : false;
			var isActive = false;
			switch (btnId) {
				case "$select":
					isActive = !isSignature;
					break;
				case "$edit":
					isActive = (!isSignature && this.nativePenAnnotatedImage);
					break;
				case "$camera":
					isActive = (!isSignature && this.nativeImageCamera);
					break;
				case "$sign":
					isActive = (isSignature && this.nativeSignatureImage);
					break;
				case "$delete":
					isActive = true;
					break;
				default:
					isActive = false;
			}
			return isActive;
		},
		_buildEditImageField: function(parentSlot, docUrl, empty) {
			// layout structure
			this.$$layout = $('<div class="s-m-img-layout">');
			// building 3 components. clear button, core and upload button. All clickable.

			var core = uiUtils.createDomElement("a", ["img-responsive", "s-m-field-img", "ctrl-evt-click"], null, {
				"href": "#"
			}, $("<section>").appendTo(this.$$layout));
			var actions = [{
				id: "$select",
				active: this._isBtnActive("$select"),
				css: "s-m-select",
				text: locale.text("image.label.selectimage")
			}, {
				id: "$edit",
				active: this._isBtnActive("$edit"),
				css: "s-m-edit " + fontUtils.getIconByName("$image-ink")
			}, {
				id: "$sign",
				active: this._isBtnActive("$sign"),
				css: "s-m-sign " + fontUtils.getIconByName("$image-sign")
			}, {
				id: "$camera",
				active: this._isBtnActive("$camera"),
				css: "s-m-camera " + fontUtils.getIconByName("$image-camera")
			}, {
				id: "$delete",
				active: this._isBtnActive("$delete"),
				css: "s-m-delete " + fontUtils.getIconByName("$delete")
			}];

			$(_getHtml("footer", {
				actions: actions
			})).appendTo(this.$$layout);

			// use default image if empty content
			if (empty) {
				core = uiUtils.createDomElement("span", [this._getDefaultIcon(), "s-m-empty", "img-responsive"], null, null, core);
			} else {
				core = this._newImg(core, docUrl);
				this.img = core;
			}
			parentSlot.appendChild(this.$$layout[0]);

			var inputDom = uiUtils.createDomElement("input", ["ctrl-evt-change"], null, {
				"type": "file",
				"accept": "image/*"
			}, parentSlot);
			this.$$elmt.find(".s-m-img-layout > footer > a[action='$delete']").toggle(!empty);
			this.$$elmt.find(".s-m-img-layout > footer > a[action='$edit']").toggle(!empty);
		},
		onClick: function(evt) {
			var dom = evt.target.nodeName.toLowerCase() === "span" ? evt.target.parentNode : evt.target;
			var parentDom;
			// detail mode
			if (!this.$isEditMode && dom.nodeName.toLowerCase() === "a") {
				parentDom = dom;
			}
			// edit mode
			else {
				parentDom = dom.parentNode;
			}
			if (!parentDom) return;
			var hdl = _handlersMap[$(dom).attr("action") || parentDom.nodeName.toLowerCase()];
			if (!hdl) return;
			this[hdl](evt, dom);
		},
		_getBinaryContent: function() {
			var binary = this.getDao().getValue(this.$bind + "-LOCALURL");
			if (!binary && this.img) {
				var canvas = document.createElement('canvas');
				var ctx = canvas.getContext("2d");
				ctx.canvas.width = this.img.naturalWidth;
				ctx.canvas.height = this.img.naturalHeight;
				ctx.drawImage(this.img, 0, 0, this.img.naturalWidth, this.img.naturalHeight);
				binary = canvas.toDataURL();
			}
			return binary;
		},
		/*
			When clear button is clicked
			- Removes  content
			- Hide clear button
			- Update upload button text if image
		*/
		_onClearClick: function(evt, dom) {
			var clearBtn = dom;
			// get core elmt (the anchor containing the image or the file)
			var core = this.$$elmt.find(".s-m-img-layout > section > a")[0];
			// remove image/doc
			uiUtils.empty(core);
			// update image/doc
			core = uiUtils.createDomElement("span", [this._getDefaultIcon(), "s-m-empty"], null, null, core);
			// update dao
			this.setValue(null);
			var element = this.$$elmt.find(".s-m-value > input");
			//element.val(''); // to trigger onchange
			element.replaceWith(element = element.clone(true)); // element.val('') does not work with IE
			$(clearBtn).toggle();
			if (this._isBtnActive('$edit')) {
				var editBtn = this.$$elmt.find(".s-m-img-layout > footer > a[action='$edit']")[0];
				$(editBtn).toggle();
			}
			// update upload button
			this._setLocalUrlValue(null);
		},
		_onEditClick: function(evt, dom) {
			var self = this;
			var editBtn = this.$$elmt.find(".s-m-img-layout > footer > a[action='$edit']")[0];
			$(editBtn).toggle();
			native.getModule("penAnnotatedImage").getAnnotatedImage({
				title: self.getTitle(),
				imageData: self._getBinaryContent()
			}).then(
				function(data) {
					if (data && data.action === "ok") {
						if (data.imageData && data.imageData.length) {
							// No need to scale here
							// The image is scaled before adding it and making inking available
							self.setValue(self._getImageDataValue(data.imageData));
							self._setImage(data.imageData);
						}
					} else {
						$(editBtn).toggle();
					};
				}
			);
		},
		_onSignClick: function(evt, dom) {
			var self = this;
			var signBtn = this.$$elmt.find(".s-m-img-layout > footer > a[action='$sign']")[0];
			$(signBtn).toggle();
			native.getModule("penAnnotatedImage").getSignatureImage({
				title: self.getTitle(),
				watermark: ""
			}).then(
				function(data) {
					if (data && data.action === "ok") {
						if (data.imageData && data.imageData.length) {
							// We do not scale images that are input as a signature since
							// signature images created by the wrapper have a defined size already 
							self.setValue(self._getImageDataValue(data.imageData));
							self._setImage(data.imageData);
						}
					}
					$(signBtn).toggle();
				}
			);
		},
		_onCameraClick: function(evt, dom) {
			var self = this;
			var signBtn = this.$$elmt.find(".s-m-img-layout > footer > a[action='$camera']")[0];
			$(signBtn).toggle();
			native.getModule("camera").getPictureFromCamera({})
				.then(function(data) {
					if (data && data.action === "ok") {
						if (data.imageData && data.imageData.length > 0) {
							return self._checkScale(data.imageData);
						}
					}
					return null;
				})
				.then(function(imageData) {
					if (imageData) {
						self.setValue(self._getImageDataValue(imageData));
						self._setImage(imageData);
					}
				})
				.always(function() {
					$(signBtn).toggle();
				});
		},
		/* 
		When upload button is clicked (or empty image), run upload and update dao
	*/
		_onUploadClick: function(evt, dom) {
			var $$inputDom = this.$$elmt.find(".s-m-value > input");
			try {
				$$inputDom.trigger("click");
			} catch (e) {
				//In case of wrapper and close the dialog box without any file selected, jquery catch 
				//on statement "Prevent re-triggering of the same event" (jQuery 2.1.1 line 4357) 
				//and does not reinit the property "triggered".
				// So we do it in place of jquery !
				jQuery.event.triggered = undefined;
			}
		},
		/*
		Runs image selection within the client device
		Updates dao
	*/
		onChange: function(evt) {
			var self = this;

			if (evt.target.nodeName.toLowerCase() === 'input') {
				var files = evt.target.files; // FileList object
				var res = {};
				if (files && files.length) {
					var f = files[0];
					if (!f.type.match('image.*')) return;
					res.$contentType = f.type;
					res.$fileName = f.name;
					res.$type = "image";

					var reader = new FileReader();
					reader.onload = (function(theFile, cont) {
						return function(e) {
							self._checkScale(e.target.result).then(function(src) {
								cont.$value = self._getImageDataValue(src);
								delete cont.$url;
								self.setValue(cont.$value);
								self.setFocus();
								self._setImage(src);
							});
						};
					})(f, res);
					reader.readAsDataURL(f);

				}
			}
		},
		_getImageDataValue: function(content) {
			var i = content.indexOf("base64,");
			if (i > 0) {
				return content.substring(i + 7);
			} else {
				return content.result;
			}
		},
		_setLocalUrlValue: function(src) {
			this.getDao().setValue(this.$bind + "-LOCALURL", src);
		},
		_getLocalUrlValue: function() {
			return this.getDao().getValue(this.$bind + "-LOCALURL");
		},
		/*
		Set image based on client image src
		*/
		_setImage: function(src) {
			this._setLocalUrlValue(src);
			// get core elmt (the anchor containing the image)
			var core = this.$$elmt.find(".s-m-img-layout > section > a")[0];
			// remove image
			uiUtils.empty(core);
			// update image
			core = this._newImg(core, src);
			// toggle clearBtn
			this.$$elmt.find(".s-m-img-layout > footer > a[action='$delete']").toggle(true);
			if (this._isBtnActive('$edit')) {
				this.$$elmt.find(".s-m-img-layout > footer > a[action='$edit']").toggle(true);
			}
		},

		/*
		 *  src: data:image/png;base64,iVBORw0KGgoAAAAN
		 */
		_checkScale: function(src) {
			if (!src) return $.smResolve(src);

			// If no authoring done or value is noscale, return source image like:
			// return $.smResolve(src);

			var scaler = native.getModule("imageScale");
			if (!scaler) return $.smResolve(src);

			var scaleMap = {
				small: {
					width: 640,
					height: 480
				},
				medium: {
					width: 1024,
					height: 768
				},
				large: {
					width: 1920,
					height: 1440
				}
			};
			var scale = scaleMap[(this.getAuthoring("$imgScale") || "original")];
			if (!scale) return $.smResolve(src);

			return scaler.getScaledImage({
				width: scale.width,
				height: scale.height,
				imageData: src,
				mode: "LIMIT_MAX_WIDTH_AND_HEIGHT" // Only supported mode right now
			})
				.then(function(result) {
					// Change so return  new image
					if (result.action === "ok") {
						return result.imageData;
					}
					// No change, maybe because size was ok already
					if (result.action === "nochange") {
						return src;
					}
					// Here we got an error but we still can return the original image
					return src;
				});
		},

		_newImg: function(parent, src) {
			return this._setImgHeight(uiUtils.createDomElement("img", ["img-responsive"], null, {
				src: src
			}, parent));
		},
		_defaultImg: function(parent, css) {
			return this._setImgHeight(uiUtils.createDomElement("span", this._getDefaultIcon() + (" " + (css || "")), null, null, parent));
		},
		_setImgHeight: function(i) {
			return $(i).height(uiUtils.imgHeight2px(this.getAuthoring("$imgHeight") || "small")).get(0);
		},
		_getDefaultIcon: function() {
			var css = "s-m-icon-" + (this.article.$emptyIconSize || "small") + " ";
			return css + (this.article.$emptyImageIcon || fontUtils.getEmptyImageDefIcon()).css;
		}
	});

exports.Klass = _Klass;
