// syracuse-tablet/html/js/controls/field/ctrlText.js
"use strict";
var utils = require('syracuse-tablet/html/js/helpers/utils');
var Base = require('syracuse-tablet/html/js/controls/field/ctrlFieldBase').Klass;
var uiUtils = require('syracuse-tablet/html/js/ui/uiUtils');
var escaper = require('syracuse-tablet/html/js/helpers/html-escape/index').escaper({
	warn: console.warn.bind(console),
});
var rtf = require('syracuse-tablet/html/js/helpers/syracuse-rtf/index');

var _Klass = utils.defineClass(

	function(controller, article, prototype, options) {
		var self = this;
		Base.call(self, controller, article, prototype, options);
	}, Base, {
		buildFieldValue: function(parentSlot, buildOptions) {
			var self = this;
			var value = self.getFormattedValue(self.getValue());

			// Create the enriched text content under the standard s-m-value parent
			var parent = Base.prototype.buildFieldValue.call(self, parentSlot, buildOptions);

			// enriched text container div
			var textContainer = uiUtils.createDomElement("div", ["s-m-enriched-text", "well"], null, null, parent);

			var docUrl = self.prototype.data("$url", self.getDao()) || "";

			//docUrl += '?salt=' + ((new Date()).getTime());
			$.ajax({
				url: docUrl,
				success: function(data, textStatus, jqXHR) {
					var ss = this;
					var $$dataArr = [];
					switch (ss.$type) {
						case "text/plain":
							ss.buildFieldFromPlain(data, textContainer);
							break;
						case "text/html":
							// TODO remove || $$dataArr
							ss.buildFieldFromHtml(data, textContainer);
							break;
						case "text/rtf":
							// TODO remove || $$dataArr
							ss.buildFieldFromRtf(data, textContainer);
							break;
					}
				},
				error: function(jqXHR, textStatus, errorThrown) {
					alert("Error: Ajax failed");
				},
				context: this
			});
		},
		buildFieldFromPlain: function(data, parent) {
			$(parent).append($.parseHTML(escaper(data.replace(/\n/g, "<BR/>"))));
		},
		buildFieldFromRtf: function(data, parent) {
			this.buildFieldFromHtml(rtf.toHtml(data), parent);
		},
		buildFieldFromHtml: function(data, parent) {
			$(parent).append($.parseHTML(escaper(data)));
		}
	});

exports.Klass = _Klass;
