// syracuse-tablet/html/js/controls/field/ctrlCombo.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var uiUtils = require('syracuse-tablet/html/js/ui/uiUtils');
var Base = require('syracuse-tablet/html/js/controls/field/ctrlFieldBase').Klass;

var _Klass = utils.defineClass(

	function(controller, article, prototype) {
		var self = this;
		Base.call(self, controller, article, prototype);
	}, Base, {
		buildFieldValue: function(parentSlot, buildOptions) {
			var self = this;
			self.choiceList = self.prototype.data("$value").$enum;
			if (self.$isEditMode) {
				var $format = self.prototype.data("$format");

				// if '$format' is set it can only have values '$combo' or '$radio'
				if ($format && ($format != '$radios' && $format != '$combo')) {
					throw new Error("Choice can only have formats '$radios' or '$combo'");
				}

				// ## set display in case of edit mode ##
				var value = self.currentValue = self.getValue();
				var domValue;
				if (self.choiceList.length > 0) {
					// display in case of '$radios' format
					if (self.prototype.data("$format") == '$radios') {
						self.radioItemList = [];
						$.each(self.choiceList, function(index, choiceItem) {
							// building choice item display
							var choiceItemDom = uiUtils.createDomElement("div", ["s-m-field-radio-item"]);
							var inputDom = uiUtils.createDomElement("input", ["s-m-field-choice-radio", "ctrl-evt-click"], null, {
								"type": "radio",
								"value": index,
								"name": "s-m-radio-" + self.id
							});
							if (value && choiceItem.$value == value) {
								inputDom.setAttribute("checked", true);
							}
							var titleDom = uiUtils.createDomElement("div", ["s-m-field-choice-radio-title"], choiceItem.$title);

							// build radioItemList array
							self.radioItemList.push({
								"inputDom": inputDom,
								"titleDom": titleDom
							});

							// appending to parent layout
							choiceItemDom.appendChild(inputDom);
							choiceItemDom.appendChild(titleDom);
							parentSlot.appendChild(choiceItemDom);
						});
					}

					// default format is '$combo'
					else {

						var selectDom = uiUtils.createDomElement("select", "ctrl-evt-change");

						// build list
						$.each(self.choiceList, function(index, choiceItem) {
							var optDom = uiUtils.createDomElement("option", null, choiceItem.$title, {
								"value": choiceItem.$value
							}, selectDom);
							if ((choiceItem.$value == value) || (!value && index == 0)) {
								optDom.setAttribute("selected", true);
							}
						});

						parentSlot.appendChild(selectDom);
					}
				}
			} else {
				if (self.choiceList.length > 0) {
					self.createFieldValueElement(parentSlot, self._getChoiceTitle(value));
				}
			}
		},
		_getChoiceTitle: function(value) {
			var self = this;
			var res;
			$.each(self.choiceList, function(index, choice) {
				res = choice.$value == value ? choice.$title : res;
				if (res) {
					return false;
				}
			});
			return res;
		},
		_getChoiceValue: function(title) {
			var self = this;
			var res;
			$.each(self.choiceList, function(index, choice) {
				res = choice.$title == title ? choice.$value : res;
				if (res) {
					return false;
				}
			});
			return res;
		},
		setDisplayValue: function(value, index, pickerText) {
			var self = this;

			if (!self.$isEditMode) {
				self.fieldValue.textContent = self._getChoiceTitle(value);
			} else {
				// in case of radio choice
				if (self.prototype.data("$format") == '$radios') {
					// apply change if new value
					if (value != self.currentValue) {
						// update widget
						$.each(self.radioItemList, function(idx, radioItem) {
							setTimeout(function() {
								// if new choice index is defined
								if (index) {
									radioItem.inputDom.checked = index == idx;
								}
								// if new choice index is not defined
								else {
									radioItem.inputDom.checked = self.choiceList[idx].$value == value;
								}
							}, 5);
						});
					}
				}

			}
		},
		setFieldValue: function(value, index, pickerText) {
			var self = this;

			self.setDisplayValue(value, index, pickerText);

			self.currentValue = value;
			self.dao.setValue(self.article.$bind, value);
		},
		onClick: function(evt) {
			var self = this;
			if (self.prototype.data("$format") == '$radios') {
				self.setFieldValue(self._getChoiceValue(evt.target.nextElementSibling.textContent), $(evt.target.parentNode).index());
			}
		},
		onChange: function(evt) {
			this.setFieldValue(evt.target.value, evt.target.selectedIndex, evt.target.selectedOptions[0].textContent);
		}
	});

exports.Klass = _Klass;
