// syracuse-tablet/html/js/controls/field/ctrlText.js
"use strict";
var utils = require('syracuse-tablet/html/js/helpers/utils');
var Base = require('syracuse-tablet/html/js/controls/field/ctrlFieldBase').Klass;
var uiUtils = require('syracuse-tablet/html/js/ui/uiUtils');
var escaper = require('syracuse-tablet/html/js/helpers/html-escape/index').escaper({
	warn: console.warn.bind(console),
	allowedAttributes: ["style"],
	allowedTags: ["span"]
});
var ajax = require('syracuse-tablet/html/js/ajax/ajax').getAjax().ajax;
var rtf = require('syracuse-tablet/html/js/helpers/syracuse-rtf/index');

var _defHeight = 50;

/**
 * Text field
 * authoring:
 * 		$height	set the height of the image - default is _defHeight
 */
var _Klass = utils.defineClass(

	function CtrlText(controller, article, prototype, options) {
		var self = this;
		Base.call(self, controller, article, prototype, options);
		self.cssType = "s-m-clob";
	}, Base, {
		buildFieldValue: function(parentSlot, buildOptions) {
			var self = this;
			var value = self.getFormattedValue(self.getValue());
			// Create the enriched text content under the standard s-m-value parent
			var parent = Base.prototype.buildFieldValue.call(self, parentSlot, buildOptions);
			// enriched text container div
			var $$container = $(uiUtils.createDomElement("div", ["s-m-enriched-text"], null, null, parent));
			var h = parseInt(self.getAuthoring("$height") || _defHeight, 10);
			$$container.css = ({
				height: h + "px"
			});
			var docUrl = self.prototype.data("$url", self.getDao()) || "";
			ajax("GET", docUrl, null, null, {
				noJsonParsing: true
			}).then(function(result) {
				var responseText = result.responseText || "";
				var $$dataArr = [];
				switch (self.$type) {
					case "text/plain":
						self.buildFieldFromPlain(responseText, $$container);
						break;
					case "text/html":
						self.buildFieldFromHtml(responseText, $$container);
						break;
					case "text/rtf":
						self.buildFieldFromRtf(responseText, $$container);
						break;
				}
			}).fail(function(e) {
				self.insertError(e.message);
				self.log(e);
			});
		},
		/**
		 * Display an error
		 */
		insertError: function(html) {
			var err = ['<div class="s-m-error">'];
			err.push(html);
			err.push("</div>");
			$(err.join('')).appendTo(this.$$elmt);
		},
		buildFieldFromPlain: function(data, $$parent) {
			$$parent.append($.parseHTML(escaper(data.replace(/\n/g, "<BR/>"))));
		},
		buildFieldFromRtf: function(data, $$parent) {
			this.buildFieldFromHtml(rtf.toHtml(data), $$parent);
		},
		buildFieldFromHtml: function(data, $$parent) {
			$$parent.append($.parseHTML(escaper(data)));
		}
	});

exports.Klass = _Klass;
