// syracuse-tablet/html/js/controls/field/ctrlImage.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var fontUtils = require('syracuse-tablet/html/js/ui/fontUtils');
var Base = require('syracuse-tablet/html/js/controls/field/ctrlFieldBase').Klass;
var uiUtils = require('syracuse-tablet/html/js/ui/uiUtils');
var locale = require('syracuse-tablet/html/js/helpers/locale');

var _handlersMap = {
	"header": "_onClearClick",
	"footer": "_onUploadClick",
};

/**
 * Image  field
 * authoring:
 * 		$imgHeight	set the height of the image - xsmall, small...
 */
var _Klass = utils.defineClass(

	function CtrlImage(controller, article, prototype, options) {
		Base.call(this, controller, article, prototype, options);
		this.cssType = "s-m-img";
	}, Base, {
		initReuseProperties: function() {
			Base.prototype.initReuseProperties.call(this);
			this.$$layout = null;
		},
		applyMetaData: function(metaData) {
			Base.prototype.applyMetaData.call(this, metaData);
			if (this.$isEditMode && this.$$layout) {
				var disabled = metaData ? metaData.$isReadOnly || metaData.$isDisabled : false;
				this.$$layout.children("header").toggle(disabled !== true);
				this.$$layout.children("footer").toggle(disabled !== true);
			}
		},

		buildFieldValue: function(parentSlot, buildOptions) {
			if (!this.article.$imgHeight) {
				this.article.$imgHeight = buildOptions && (buildOptions.displayCtx === "table" || buildOptions.displayCtx === "card") ? "small" : "medium";
			}
			var value = this.getFormattedValue(this.getValue());
			// Create the image stuff under the standard s-m-value parent
			var imgParent = Base.prototype.buildFieldValue.call(this, parentSlot, buildOptions);
			var docUrl = this._getLocalUrlValue();
			if (!docUrl) {
				docUrl = this.prototype.data("$url", this.getDao()) || "";
				//TODO use $etag or else - No need to force reload if updateLayout - data are in the browser's cache
				if (buildOptions.updateLayout !== true) {
					docUrl += '?salt=' + ((new Date()).getTime());
				}
			}
			var empty = value == null || value === "" || docUrl.length === 0;
			if (empty) $(imgParent).addClass("s-m-empty");
			// if FileReader not supported by the browser, we display detail view (fix for Safari)
			if (this.$isEditMode && typeof FileReader !== "undefined") {
				this._buildEditImageField(imgParent, docUrl, empty);
			} else {
				if (this.article.$imgDisplayIcon === true) {
					// We want to display images as icons - enabled only for arrays in table mode
					var css;
					if (empty) {
						css = fontUtils.getColImageDefIcon(true).css;
					} else {
						css = (this.article.$imgIcon || fontUtils.getColImageDefIcon(false)).css;
					}
					return $(imgParent).addClass(css);
				} else if (empty) {
					// Only icon
					if (buildOptions && buildOptions.displayCtx === "card") {
						// same height for all cards -> We set the height
						this._defaultImg(imgParent, null);
					} else {
						$(imgParent).addClass(this._getDefaultIcon());
					}
				} else {
					// link - Maybe we could manage a modal to display the image 
					var e = uiUtils.createDomElement("a", ["img-responsive ctrl-evt-click"], null, {
						"href": "#"
					}, imgParent);
					// Image with link
					e = this._newImg(e, docUrl);
				}
			}
		},
		_buildEditImageField: function(parentSlot, docUrl, empty) {
			// layout structure
			this.$$layout = $('<div class="s-m-img-layout">');
			// building 3 components. clear button, core and upload button. All clickable.
			var clearBtn, uploadBtn, core;
			clearBtn = uiUtils.createDomElement("a", ["ctrl-evt-click"], null, {
				"href": "#"
			}, $("<header>").appendTo(this.$$layout));
			var classCss = ["s-m-clear"];
			classCss.push(fontUtils.getIconByName("$delete"));
			clearBtn = uiUtils.createDomElement("span", classCss, null, null, clearBtn);
			$(clearBtn).toggle(!empty);

			core = uiUtils.createDomElement("a", ["img-responsive", "s-m-field-img", "ctrl-evt-click"], null, {
				"href": "#"
			}, $("<section>").appendTo(this.$$layout));

			uploadBtn = uiUtils.createDomElement("a", ["ctrl-evt-click"], locale.text("image.label." + (empty ? "addimage" : "changeimage")), {
				"href": "#"
			}, $("<footer>").appendTo(this.$$layout));
			// use default image if empty content
			if (empty) {
				core = uiUtils.createDomElement("span", [this._getDefaultIcon(), "s-m-empty", "img-responsive"], null, null, core);
			} else {
				core = this._newImg(core, docUrl);
			}
			parentSlot.appendChild(this.$$layout[0]);

			var inputDom = uiUtils.createDomElement("input", ["ctrl-evt-change"], null, {
				"type": "file",
				"accept": "image/*"
			}, parentSlot);
		},
		onClick: function(evt) {
			var dom = evt.target.nodeName.toLowerCase() === "span" ? evt.target.parentNode : evt.target;
			var parentDom;
			// detail mode
			if (!this.$isEditMode && dom.nodeName.toLowerCase() === "a") {
				parentDom = dom;
			}
			// edit mode
			else {
				parentDom = dom.parentNode;
			}
			if (!parentDom) return;
			var hdl = _handlersMap[parentDom.nodeName.toLowerCase()];
			if (!hdl) return;
			this[hdl](evt, dom);
		},
		/*
			When clear button is clicked
			- Removes  content
			- Hide clear button
			- Update upload button text if image
		*/
		_onClearClick: function(evt, dom) {
			var clearBtn = dom;
			// get core elmt (the anchor containing the image or the file)
			var core = this.$$elmt.find(".s-m-img-layout > section > a")[0];
			// remove image/doc
			uiUtils.empty(core);
			// update image/doc
			core = uiUtils.createDomElement("span", [this._getDefaultIcon(), "s-m-empty"], null, null, core);
			// update dao
			this.setValue(null);
			$(clearBtn).toggle();
			// update upload button
			this.$$elmt.find(".s-m-img-layout > footer > a").text(locale.text("image.label.addimage"));
			this._setLocalUrlValue(null);
		},
		/* 
		When upload button is clicked (or empty image), run upload and update dao
	*/
		_onUploadClick: function(evt, dom) {
			var $$inputDom = this.$$elmt.find(".s-m-value > input");
			$$inputDom.trigger("click");
		},
		/*
		Runs image selection within the client device
		Updates dao
	*/
		onChange: function(evt) {
			var self = this;

			if (evt.target.nodeName.toLowerCase() === 'input') {
				var files = evt.target.files; // FileList object
				var res = {};
				if (files && files.length) {
					var f = files[0];
					if (!f.type.match('image.*')) return;
					res.$contentType = f.type;
					res.$fileName = f.name;
					res.$type = "image";

					var reader = new FileReader();
					reader.onload = (function(theFile, cont) {
						return function(e) {
							var i = e.target.result.indexOf("base64,");
							if (i > 0) {
								cont.$value = e.target.result.substring(i + 7);
							} else {
								cont.$value = e.target.result;
							}
							delete cont.$url;
							self.setValue(cont.$value, true);
							self._setImage(e.target.result);
						};
					})(f, res);
					reader.readAsDataURL(f);

				}
			}
		},
		_setLocalUrlValue: function(src) {
			this.getDao().setValue(this.$bind + "-LOCALURL", src);
		},
		_getLocalUrlValue: function() {
			return this.getDao().getValue(this.$bind + "-LOCALURL");
		},
		/*
		Set image based on client image src
		*/
		_setImage: function(src) {
			this._setLocalUrlValue(src);
			// get core elmt (the anchor containing the image)
			var core = this.$$elmt.find(".s-m-img-layout > section > a")[0];
			// remove image
			uiUtils.empty(core);
			// update image
			core = this._newImg(core, src);
			// update upload button
			this.$$elmt.find(".s-m-img-layout > footer > a").text(locale.text("image.label.changeimage"));
			// toggle clearBtn
			this.$$elmt.find(".s-m-img-layout > header > a").toggle(true);
		},
		_newImg: function(parent, src) {
			return this._setImgHeight(uiUtils.createDomElement("img", ["img-responsive"], null, {
				src: src
			}, parent));
		},
		_defaultImg: function(parent, css) {
			return this._setImgHeight(uiUtils.createDomElement("span", this._getDefaultIcon() + (" " + (css || "")), null, null, parent));
		},
		_setImgHeight: function(i) {
			return $(i).height(uiUtils.imgHeight2px(this.getAuthoring("$imgHeight") || "small")).get(0);
		},
		_getDefaultIcon: function() {
			var css = "s-m-icon-" + (this.article.$emptyIconSize || "small") + " ";
			return css + (this.article.$emptyImageIcon || fontUtils.getEmptyImageDefIcon()).css;
		}
	});

exports.Klass = _Klass;
