// syracuse-tablet/html/js/application/application.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var modal = require('syracuse-tablet/html/js/ui/modal');
var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("app");
var notifications = require('syracuse-tablet/html/js/helpers/notifications');
var globals = require('syracuse-tablet/html/js/helpers/globals');
var auth = require('syracuse-tablet/html/js/application/authentication');
var eventListener = require('syracuse-tablet/html/js/application/eventListener');
var pageLoader = require('syracuse-tablet/html/js/application/pageLoader');
var initContext = require('syracuse-tablet/html/js/init/initContext');
var settings = require('syracuse-tablet/html/js/application/settings');
var uiutils = require('syracuse-tablet/html/js/ui/uiUtils');
var pageHelper = require('syracuse-tablet/html/js/application/pageHelper');
var authoring = require('syracuse-tablet/html/js/authoring/authoring');
/**
 * Default application config
 */
var _appConfig = {
	transition: null,
	// FDB - openLinkInVignette=true to allow test in vignettes (children page, lookup...)
	openLinkInVignette: true,
	arrays: {
		gridMaxCols: 5
	}
};
var _refreshBrowserHistory = function(app) {
	if (app.history.length == 0) {
		window.history.pushState({
			d: 3,
			r: 6
		}, "uuid", null);
	} else {
		window.history.replaceState({
			d: 3,
			r: 6
		}, "uuid", null);
	}
};
/**
 * Subscriptions
 */
var _subscribe = function(self) {
	notifications.subscribe(self, ["sm.context.changed", "sm.logout", "sm.switch.app", "sm.modal.open", "sm.modal.close"], 1);
};
/**
 * Tablet application class
 */
var _Application = utils.defineClass(
	function($$elmt, factory) {
		var self = this;
		self.$$elmt = $$elmt;
		self.currentPage = null;
		self._$config = $.extend(true, {}, _appConfig);
		self.factory = factory;
		self._pageLoader = null;
		self._nbModals = 0;
		self._wait = uiutils.waitWheelCreate($$elmt);
		// Id of the home dashboard of current application
		// Used to go to home page
		self.homePageId = null;
		// Default appid when login page is loaded without current application
		self.uuid = "root";
		// History stack
		self.history = [];
	}, null, {
		/**
		 * Called after object has been created
		 **/
		init: function() {
			var self = this;
			// Can't be called in constructor
			eventListener.bindEvents(self);
			self.dao = self.factory.createDaoApp(self);
			_subscribe(self);
			self.$$elmt.show();
		},

		destroy: function() {
			var self = this;
			if (self._wait) uiutils.waitWheelDestroy(self._wait);
			notifications.unsubscribe();
			pageHelper.removeAllPages(self.$$elmt);
			self.currentPage = null;
			if (self._pageLoader == null) {
				self._pageLoader.destroy();
				self._pageLoader = null;
			}
			utils.unbindObj(self);
			utils.unbindObj(self.$$elmt);
			$(window).unbind();
		},
		/**
		 * True if current 'mobile application' is welcome application
		 */
		isWelcomeApplication: function() {
			return this.uuid === "$welcomeApplication";
		},
		/**
		 * For page display
		 */
		getTitle: function() {
			return this.dao.getApplicationName();
		},
		/**
		 * Goto home page of current 'mobile application'
		 */
		gotoHome: function() {
			var self = this;
			setTimeout(function() {
				self.changePage(self.homePageId);
			});
		},
		/**
		 * Accept path like "array.gridMaxCols"
		 */
		$config: function(id) {
			return utils.getPropByPath(this._$config, id, this._$config);
		},
		/**
		 * User, role, endpoint or language changed
		 * !! resolve/reject deferred is not null
		 * -> deferred used to catch error in transition phase - before page has been loaded
		 */
		notifContextChanged: function(deferred) {
			this.gotoWelcomeApplication(deferred);
		},

		notifLogout: function() {
			log && log("Application Notification Logout");
			globals.setUserCtx(null);
		},

		notifModalOpen: function() {
			this._nbModals++;
			// console.log("notifModalOpen nbModals=" + this._nbModals)	
		},

		notifModalClose: function() {
			this._nbModals = Math.max(this._nbModals - 1, 0);
			// console.log("notifModalClose nbModals=" + this._nbModals)
		},

		hasModalOpen: function() {
			//  console.log("_nbModals=" + this._nbModals)
			return this._nbModals > 0;
		},

		_actGotoWelcomeApplication: function() {
			return this.gotoWelcomeApplication();
		},
		/**
		 * Called by link with data-action = historyBack
		 */
		_actHistoryBack: function() {
			window.history.back();
		},
		// Switch role, endpoint, language
		_actSwitchContext: function() {
			var self = this;
			initContext.init(true, true).then(function(context) {}).fail(function(e) {
				modal.error("Swicth context error", e);
			});
		},
		_actClearCache: function() {
			var self = this;
			var deferred = $.Deferred();
			modal.confirm("clearCache", function(confirmed) {
				if (confirmed) {
					self.waitWheelStart();
					globals.getCache().clearCache().then(function(e) {
						self.waitWheelStop();
						deferred.resolve(null);
					}).fail(function(e) {
						setTimeout(function() {
							utils.modalError("Clear cache error", e, function() {
								self.waitWheelStop();
								deferred.resolve(null);
							});
						}, 10);
					});
				} else {
					deferred.resolve(null);
				}
			});
			return deferred.promise();
		},
		_actDesignPage: function() {
			authoring.startDesignPage();
		},
		/**
		 * User has selected a new Syracuse mobile application
		 * appInfo json description of application
		 */
		notifSwitchApp: function(appName) {
			this._setCurrentApplication(globals.getApplicationJson(appName));
		},

		/**
		 * Got o home page of tablet applications - Welcome dashboard
		 * deferred to resolve/reject in not null - used by asynchronous calls
		 * Ex: For login page we display the error in login page instead of modal
		 */
		gotoWelcomeApplication: function(deferred) {
			return this._setCurrentApplication(globals.getWelcomeAppJson(), deferred);
		},
		/**
		 * Set the current 'mobile application'
		 * Called after a switch context (application selection) or when tablet application is launched
		 * 'Welcome dashboard' page is a real application with only one dashboard
		 *  deferred param :
		 *  	deferred to resolve/reject if not null - used by asynchronous calls to be notified
		 */
		_setCurrentApplication: function(appInfo, deferred) {
			var self = this;
			var _fail = function(msg, e) {
				self.waitWheelStop();
				if (deferred) {
					deferred.reject(e);
				} else {
					modal.error(msg, e, function() {
						if (self.currentPage) self.currentPage.activate();
					});
				}
			};
			var _success = function(appDetail) {
				try {
					self.waitWheelStop();
					if (!appDetail.$application.$uuid) throw new Error("No application $uuid - " + appDetail.applicationName);
					if (self.uuid !== appDetail.$application.$uuid) {
						// Application changed
						self.dao.setApplication(appDetail);
						self.homePageId = self.dao.getHomeDashboard();
						self.uuid = appDetail.$application.$uuid;
					}
					self.gotoHome();
					if (deferred) deferred.resolve();
				} catch (e) {
					_fail("Error loading new application", e);
				}
			};
			try {
				if (!appInfo) throw new Error("Null appInfo");
				log && log("Switch to Syracuse Mobile Application " + appDetail.applicationName);
				if (self.currentPage) {
					self.currentPage.deactivate();
				}
				self.waitWheelStart();
				settings.getMyApplicationDetail(appInfo).then(function(appDetail) {
					_success(appDetail);
				}).fail(function(e) {
					_fail("Error loading application", e);
				});
			} catch (e) {
				_fail("Error switching application", e);
			}
		},
		/**
		 * Open a page
		 * !! Do not call directly - Use changePage
		 */
		_gotoPage: function(state, back) {
			var self = this;
			if (!self._pageLoader) {
				/**
				 * Page loader shared with vignette
				 * Load a page
				 */
				self._pageLoader = new pageLoader.Klass({
					waitStop: function() {
						self.waitWheelStop();
					},
					getRootElmt: function() {
						return self.$$elmt;
					},
					waitStart: function() {
						self.waitWheelStart();
					},
					getCurrentPage: function() {
						return self.currentPage;
					},
					setCurrentPage: function(page) {
						self.currentPage = page;
						if (self.isWelcomeApplication()) {
							self.historyReset(page);
						} else if (page.state.name === self.homePageId) {
							// Destroy all cached pages except current home page
							console.log("Home page " + self.homePageId + " - Destroy all pages except current");
							pageHelper.removeAllPages(self.$$elmt, page);
						}
					},
					historyPush: function(state) {
						// Store page in browser history to enable back - accepts only json objects
						console.log("pushState before", state.name, "\thistory=" + self.history.length, "\twinHistory=" + window.history.length);
						_refreshBrowserHistory(self);
						self.history.push(state);
						console.log("pushState after", state.name, "\thistory=" + self.history.length, "\twinHistory=" + window.history.length);
					},
					succeeded: function(dstPage, parentNotifInfo) {
						// Call resize handler each time a page has been loaded for calculations
						globals.triggerResize();
						// call on page resize for calculation
						if (parentNotifInfo && parentNotifInfo.notifId) {
							/**
							 * If the child application wants to notify the parent it should return info for notification
							 * parentNotifInfo contain :
							 * 		notifId:  	with the notification id
							 * 		notifData: 	with data  needed for processing this event
							 * 		controlId: 	If action as been trigered by a control
							 * 		parentId: 	id of the page
							 * TODO - change the notification process to create one subscription per page to not notify inactive pages
							 * Currently we test if page is active before processing a notifications
							 */
							if (dstPage.id === parentNotifInfo.parentId) {
								notifications.publish(parentNotifInfo.notifId, parentNotifInfo.notifData, parentNotifInfo.controlId);
							} else {
								utils.modalError("Child page notification warning", "Parent page is not the current page");
							}
						}
					}
				});
			}
			return self._pageLoader.load(state, back);
		},
		historyReset: function(page) {
			var self = this;
			// Destroy all cached pages except page parameter
			pageHelper.removeAllPages(self.$$elmt, page);
			self.history = [];
		},
		/**
		 * Called by gotoPage or window.back
		 * !! never call this method directlt - use changePage
		 * Manage state stack - local or browser one depending on settings
		 */
		historyPush: function(state) {},
		/**
		 * Called by browser history.back
		 * !! never call this method directlty - Call window.back
		 */
		historyPop: function(event) {
			try {
				var self = this;
				event.preventDefault();
				// No back if a modal is open
				if (self._nbModals > 0) return;
				// Check bottom of stack
				if (self.history.length <= 1) {
					if (self.uuid === "root") {
						// No application has been loaded - Current page is first login page
						return;
					}
					// No more history - clear stack and pages and display welcome application
					self.historyReset();
					setTimeout(function() {
						self.gotoWelcomeApplication();
					}, 0);
					return;
				}
				console.log("popstate before ", "\thistory=" + self.history.length, "\twinHistory=" + window.history.length);
				// Remove current state
				self.history.pop();
				// Get previous state
				var state = self.history[self.history.length - 1];
				if (!state) throw new Error("Unexpected null history state");
				log && log("back", "start", "\n" + JSON.stringify(state, null, 2));
				//log && log("popstate", "history.length=" + window.history.length, "State=" + (state != null));
				_refreshBrowserHistory(self);
				console.log("popstate after", state.name, "\thistory=" + self.history.length, "\twinHistory=" + window.history.length);
				setTimeout(function() {
					self._gotoPage(state, true);
				}, 0);
			} catch (e) {
				modal.error("historyPop error", e);
			}
		},
		/** state - name of the page
		 * Option override standard pageInfo and allows to store a context
		 **/
		changePage: function(state, options) {
			var self = this;
			if (typeof state !== "string") throw new Error("changePage - string expected");
			return self._gotoPage(pageHelper.getPageInfo(state, options, self.currentPage));
		},

		logout: function() {
			var deferred = $.Deferred();
			var self = this;
			self.waitWheelStart();
			auth.logout().then(function(ok, message) {
				self.waitWheelStop();
				if (ok) {
					notifications.publish("sm.logout");
					self.changePage("login");
					deferred.resolve();
				} else {
					modal.error(message, function() {
						self.changePage("login");
						deferred.resolve();
					});
				}
			}).fail(function(e) {
				self.waitWheelStop();
				modal.error("Logout error", e, function() {
					self.changePage("login");
					deferred.resolve();
				});
			});
			return deferred.promise();
		},

		/**
		 * Refresh current page - called from eventListener
		 */
		refreshPage: function(options) {
			var self = this;
			if (!self.currentPage) throw new Error("No current page");
			self.currentPage.refresh(options).then(function() {
				//	Eventually do something in vignette control	
			}).fail(function(e) {
				self.modalError("Refresh page", e);
			});
		},
		/**
		 * Resize handler called by window.resize
		 */
		onResize: function(evt) {
			if (this.currentPage) {
				this.currentPage.onResize(evt);
			}
		},
		/**
		 * One waiting plugin per page
		 */
		waitWheelStart: function() {
			uiutils.waitWheelStart(this._wait);
		},
		waitWheelStop: function() {
			uiutils.waitWheelStop(this._wait);
		},
		/**
		 * Also exported in globals
		 */
		modalError: function(title, e, cb) {
			modal.error(title, e, cb);
		}
	});


exports.Application = _Application;
