// syracuse-tablet/html/js/application/daoApp.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');

/**
 * Class that manage access to application DATA
 *  	appli: 		Application object
 * !!use DAO methods instead of dealing directly with JSON
 */
var _Dao = utils.defineClass(
	function(appli) {
		var self = this;
		self._appli = appli;
		self.appMetaData = null;
	}, null, {

		destroy: function() {
			var self = this;
			self._appli = null;
		},
		/**
		 *Returns the JSON description of application
		 */
		isApplicationLoaded: function() {
			var self = this;
			return (self.appMetaData != null);
		},

		/**
		 * Set the JSON description of application
		 */
		setApplication: function(appMetaData) {
			var self = this;
			if (appMetaData) {
				self.appMetaData = $.extend(true, {}, appMetaData);
			}
		},

		getApplicationName: function() {
			var self = this;
			return self.appMetaData && self.appMetaData.$application.applicationName || "";
		},

		getApplicationTitle: function() {
			var self = this;
			return (self.appMetaData && (self.appMetaData.$application.title || self.appMetaData.$application.applicationName)) || "";
		},

		/*
		 * Return name of home dashboard of this application or null if not available
		 */
		getHomeDashboardName: function() {
			var self = this;
			var homeId;
			if (self.appMetaData && self.appMetaData.$dashboards) {
				for (var p in self.appMetaData.$dashboards) {
					var data = self.appMetaData.$dashboards[p];
					if (data.$dashboardName === self.appMetaData.$application.$homeDashboard.dashboardName) {
						homeId = data.$dashboardName;
					}
				}
			}
			return homeId;
		},
	});

exports.Dao = _Dao;
