// syracuse-tablet/html/js/application/daoApp.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var locale = require('syracuse-tablet/html/js/helpers/locale');

/**
 * Class that manage access to application DATA
 *  	appli: 		Application object
 * !!use DAO methods instead of dealing directly with JSON
 */
var _Dao = utils.defineClass(
	function(appli) {
		var self = this;
		self._appli = appli;
		self.appMetaData = null;
	}, null, {

		destroy: function() {
			var self = this;
			self._appli = null;
		},
		/**
		 *Returns the JSON description of application
		 */
		isApplicationLoaded: function() {
			var self = this;
			return (self.appMetaData != null);
		},

		/**
		 * Set the JSON description of application
		 */
		setApplication: function(appMetaData) {
			var self = this;
			if (appMetaData) {
				self.appMetaData = $.extend(true, {}, appMetaData);
			}
		},

		getApplicationName: function() {
			var self = this;
			return self.appMetaData && self.appMetaData.$application.applicationName || "";
		},

		getApplicationTitle: function() {
			var self = this;
			return (self.appMetaData && (self.appMetaData.$application.title || self.appMetaData.$application.applicationName)) || "";
		},

		getApplicationIcon: function() {
			return this.appMetaData && this.appMetaData.$application.iconName;
		},
		isSageApplication: function() {
			return this.appMetaData != null && this.appMetaData.$application.isSageApplication === true;
		},
		getAppBackGroundClass: function() {
			// No background class - we removed the background image
			return "";
		},
		/*
		 * Return name of home dashboard of this application or null if not available
		 */
		getHomeDashboardName: function() {
			var self = this;
			var homeDash = (self.appMetaData.$application.$homeDashboard.dashboardName || "").trim();
			if (homeDash.length == 0) {
				new Error(locale.text("error.page.homenotdefined", [self.getApplicationName()]));
			}
			if (self.appMetaData && self.appMetaData.$dashboards) {
				for (var p in self.appMetaData.$dashboards) {
					var data = self.appMetaData.$dashboards[p];
					if (data.$dashboardName === homeDash) {
						return data.$dashboardName;
					}
				}
			}
			throw new Error(locale.text("error.page.homenotfound", [self.getApplicationName(), homeDash]));
		},
	});

exports.Dao = _Dao;
