// syracuse-tablet/html/js/application/pageRegistry.js
"use strict";

var globals = require('syracuse-tablet/html/js/helpers/globals');
var utils = require('syracuse-tablet/html/js/helpers/utils');
var clientContract = require('syracuse-tablet/html/js/sdata/entities/clientContract');
var Gadget = require('syracuse-tablet/html/js/application/gadget').Klass;

var _staticPages = {};

exports.getPageInfo = function(pageName, options, currentPage) {
	return $.smResolve()
		.then(function() {
			// Check if a static page is requested first
			var pageInfo = _staticPages[pageName];
			return pageInfo;
		})
		.then(function(pageInfo) {
			// Check for client side page definition
			if (!pageInfo) {
				pageInfo = clientContract.getPageInfo(pageName, options);
			}
			return pageInfo;
		})
		.then(function(pageInfo) {
			if (pageName === "rowdetail") {
				return _getPageInfoRowDetail(pageName, options, currentPage);
			}
			// Check for pages defined by "real" mobile applications
			if (!pageInfo) {
				pageInfo = globals.getMetaData().getPageInfo(pageName);
			}
			return pageInfo;
		})
		.then(function(pageInfo) {
			if (!pageInfo) {
				// Workaround for all that is lazy loaded
				pageInfo = _createDummyPageInfo(pageName);
			} else {
				_addAuthoringInfo(pageInfo);
			}
			return pageInfo;
		})
		.then(function(pageInfo) {
			_addOptions(pageName, pageInfo, options, currentPage);

			pageInfo.uuid = utils.readableuid("page", pageInfo.type, pageInfo.name);
			return pageInfo;
		});
};

function _getPageInfoRowDetail(pageName, options, currentPage) {
	if (!currentPage) {
		throw new Error("_getRowdetailPage - Not null currentPage expected");
	}
	if (!currentPage.getControl) {
		throw new Error("_getRowdetailPage - Unexpected currentPage object - getControl method not found");
	}
	options = options || {};
	var ctrlArray = currentPage.getControl(options.controlId);
	if (!ctrlArray) {
		throw new Error("_getRowdetailPage - Array not found - id:" + options.controlId);
	}

	// Page name used to store the page - current page name + path of the selected row
	var pageName = ctrlArray.rowDetailGetPageName(currentPage.state.name);
	var pageInfo = {
		name: pageName,
		home: false,
		type: "rowdetail",
		subtype: currentPage.isEditMode() ? "edit" : "details",
		cached: false,
		refreshed: false,
		changeHash: false,
		rowDetailPrototype: ctrlArray.rowDetailGetPrototype()
	};
	return pageInfo;
}

function _addOptions(pageName, pageInfo, options, currentPage) {
	options = options || {};
	pageInfo.options = $.extend(true, {}, options);
	_addGadgetData(pageInfo, currentPage);
}

function _addGadgetData(pageInfo, currentPage) {
	var gadgetId = pageInfo.options["gadget-id"];
	if (gadgetId) {
		var prop = currentPage && currentPage.prototype && currentPage.prototype.property(gadgetId);
		var json = prop && prop.$gadget;
		pageInfo.options.gadget = json ? new Gadget(json) : null;
	}
}

function _createDummyPageInfo(pageName) {
	var pageInfo = {
		name: pageName,
		home: false,
		type: "regular",
		subtype: "",
		cached: false,
		refreshed: false,
		changeHash: false,
	};
	return pageInfo;
}

function _addAuthoringInfo(pageInfo) {
	if (!pageInfo.authoringName) {
		// Compute page name used by authoring
		var segs = pageInfo.name.split(".");
		var authoringName;
		if (segs.length === 5) {
			// remove dataset (e.g. SUPERV, ...)
			authoringName = segs[0] + "." + segs[1] + "." + segs[3] + "." + segs[4];
		} else {
			authoringName = pageInfo.name;
		}
		if (segs.length === 2) {
			authoringName = "syracuse.collaboration." + authoringName;
		}
		pageInfo.authoringName = authoringName;
	}
}

/**
 * Static preloaded pages
 * Pages belongs to all mobile applications
 */
function _registerStaticPages() {
	["login", "about"].forEach(function(name) {

		var info = {
			name: name,
			type: "static",
			subtype: "html",

			cached: false,
			refreshed: false,
			changeHash: false,

			home: false,
			disableAuthoring: true,
			isLogin: name === "login",

			/**
			 * Html have an empty proto and are base on a stacked layout for compatibility with other pages
			 * Html is appended to layout root elmt
			 * We add the proto/article in state for convenience - it can be enhanced according to the needs
			 * Currently we need only static html pages
			 */
			$prototype: {
				$properties: {}
			},
			$article: {
				$layoutType: "stack",
				$items: []
			}
		};
		_staticPages[name] = info;
	});
}

// Called only once to register static html pages
_registerStaticPages();
