// syracuse-tablet/html/js/application/settings.js
"use strict";

/*
 * This module should be used to persist user specific settings
 */

var globals = require('syracuse-tablet/html/js/helpers/globals');
var storageModule = require('syracuse-tablet/html/js/storage/storage');
var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("settings");
var voice = require('syracuse-tablet/html/js/helpers/native/nativeVoiceCommands');
var nativeApp = require('syracuse-tablet/html/js/helpers/native/nativeSageX3WUPApp');

var _currentContextUserRole = {};
var _currentContextUser = {};

var _settingsUserRole = null;
var _settingsUser = null;

/*
 *
 */
function _getContext() {
	var ctx = globals.getUserCtx();
	if (!ctx) { // In case not logged in yet!
		return {
			$user: "none",
			$role: "none",
			$lang: "en-US",
		};
	}
	return ctx;
}

/*
 *
 */
function _checkContextUserRole() {
	var ctx = _getContext();
	if (_currentContextUserRole.user != ctx.$user ||
		_currentContextUserRole.role != ctx.$role) {
		_currentContextUserRole.user = ctx.$user;
		_currentContextUserRole.role = ctx.$role;
		_currentContextUserRole.lang = ctx.$lang;
		_settingsUserRole = null;
	}
}

/*
 *
 */
function _readSettingsUserRole() {
	var storage = globals.getStorage();

	var ctx = {
		"$user": _currentContextUserRole.user,
		"$role": _currentContextUserRole.role,
		"$lang": "none"
	};

	return storage.read({
		$context: ctx,
		$collection: "$settings",
		$endpoint: "$local",
		$key: "$user_and_role"
	}).then(function(result) {
		if (result.$status === storageModule.StatusCodes.OK) {
			return result.$data;
		} else {
			return {};
		}
	});
}

/*
 *
 */
function _checkContextUser() {
	var ctx = _getContext();
	if (_currentContextUser.user != ctx.$user ||
		_currentContextUser.role != ctx.$role) {
		_currentContextUser.user = ctx.$user;
		_currentContextUser.role = ctx.$role;
		_currentContextUser.lang = ctx.$lang;
		_settingsUser = null;
	}
}

/*
 *
 */
function _readSettingsUser() {
	var storage = globals.getStorage();

	var ctx = {
		"$user": _currentContextUser.user,
		"$role": "none",
		"$lang": "none"
	};

	return storage.read({
		$context: ctx,
		$collection: "$settings",
		$endpoint: "$local",
		$key: "$user"
	}).then(function(result) {
		if (result.$status === storageModule.StatusCodes.OK) {
			_settingsUser = result.$data;
		} else {
			_settingsUser = {};
		}
		log && log("Read user settings...");
		log && log(_settingsUser);
	});
}

/*
 *
 */
function _writeSettingsUser() {
	var storage = globals.getStorage();

	var ctx = {
		"$user": _currentContextUser.user,
		"$role": "none",
		"$lang": "none"
	};

	log && log("Writing user settings...");
	log && log(_settingsUser);
	return storage.put({
		$context: ctx,
		$collection: "$settings",
		$endpoint: "$local",
		$key: "$user",
		$data: _settingsUser
	}).then(function(result) {
		if (result.$status === storageModule.StatusCodes.OK) {
			return result.$data;
		} else {
			return {};
		}
	});
}

/*
 * ======================================================
 * Pages pinned on welcome page
 * ======================================================
 */

/*
 *
 */
exports.getPinnedPages = function() {
	_checkContextUser();
	var read;
	if (_settingsUser == null) {
		read = _readSettingsUser();
	} else {
		read = $.smResolve();
	}
	return read.then(function() {
		_settingsUser.pinnedPages = _settingsUser.pinnedPages || [];
		return _settingsUser.pinnedPages;
	});
};

/*
 *
 */
exports.addPinnedPage = function(page) {
	return exports.getPinnedPages().then(function(pages) {
		if (!pages.some(function(p, idx) {
			if (p.id === page.id) {
				pages[idx] = page; // Overwrite existing
				return true;
			}
		})) {
			pages.push(page);
		}
		voice.addVoiceBookmark(page.title);
		// No return ... then, because can be done async
		_writeSettingsUser();
		return page;
	});
};

/*
 *
 */
exports.isPinnedPage = function(id) {
	return exports.getPinnedPages().then(function(pages) {
		return pages.some(function(p, idx) {
			if (p.id === id) {
				return true;
			}
		});
	});
};

/*
 *
 */
exports.removePinnedPage = function(id) {
	return exports.getPinnedPages().then(function(pages) {
		if (pages.some(function(p, idx) {
			if (p.id === id) {
				voice.removeVoiceBookmark(p.title);
				pages.splice(idx, 1); // remove and stop
				return true;
			}
		})) {
			// Only write if there was a change
			// No return ... then, because can be done async
			_writeSettingsUser();
		}
	});
};

/*
 * ======================================================
 * Apps hidden on welcome page
 * ======================================================
 */
/*
 *
 */
exports.getHiddenApplications = function() {
	_checkContextUser();
	var read;
	if (_settingsUser == null) {
		read = _readSettingsUser();
	} else {
		read = $.smResolve();
	}
	return read.then(function() {
		_settingsUser.hiddenApplications = _settingsUser.hiddenApplications || [];
		return _settingsUser.hiddenApplications;
	});
};

/*
 *
 */
exports.unhideApplication = function(applicationName) {
	return exports.getHiddenApplications().then(function(apps) {
		if (apps.some(function(app, idx) {
			if (app === applicationName) {
				apps.splice(idx, 1); // remove and stop
				return true;
			}
		})) {
			// Only write if there was a change
			// No return ... then, because can be done async
			_writeSettingsUser();
		}
	});
};

/*
 *
 */
exports.hideApplication = function(applicationName) {
	return exports.getHiddenApplications().then(function(apps) {
		if (!apps.some(function(app, idx) {
			if (app === applicationName) {
				return true;
			}
		})) {
			apps.push(applicationName);
			// No return ... then, because can be done async
			_writeSettingsUser();
		}
	});
};

/*
 * ======================================================
 * Tiles on welcome page
 * ======================================================
 */
exports.getTileConfigs = function() {
	_checkContextUser();
	var read;
	if (_settingsUser == null) {
		read = _readSettingsUser();
	} else {
		read = $.smResolve();
	}
	return read.then(function() {
		_settingsUser.tileConfigs = _settingsUser.tileConfigs || {};
		return _settingsUser.tileConfigs;
	});
};

exports.setTileConfig = function(tile) {
	return exports.getTileConfigs().then(function(configs) {
		configs[tile.tileId] = tile;
		_writeSettingsUser();
	});
};


/*
 *
 */
exports.getLastUserProfile = function(profile) {
	var ctx = {
		"$user": "none",
		"$role": "none",
		"$lang": "none"
	};
	var storage = globals.getStorage();
	return storage.read({
		$context: ctx,
		$collection: "$settings",
		$endpoint: "$local",
		$key: "$last_user_profile"
	}).then(function(result) {
		if (result.$status === storageModule.StatusCodes.OK) {
			return result.$data;
		} else {
			return null;
		}
	});
};

/*
 *
 */
exports.setLastUserProfile = function(profile) {
	var ctx = {
		"$user": "none",
		"$role": "none",
		"$lang": "none"
	};
	var storage = globals.getStorage();
	return storage.put({
		$context: ctx,
		$collection: "$settings",
		$endpoint: "$local",
		$key: "$last_user_profile",
		$data: profile
	}).then(function(result) {
		if (result.$status === storageModule.StatusCodes.OK) {
			return result.$data;
		} else {
			return {};
		}
	});
};


/*
 * These are settings that are understood to be global for the device
 * This should be things that are common for all users, roles, languages that use the current device
 * E.g. is the device type ("auto", "tablet", "smartphone")
 */
exports.getGlobalSettings = function() {
	var storage = globals.getStorage();

	var ctx = {
		"$user": "none",
		"$role": "none",
		"$lang": "none"
	};

	return storage.read({
		$context: ctx,
		$collection: "$settings",
		$endpoint: "$local",
		$key: "$global_settings"
	}).then(function(result) {
		var settings;
		if (result.$status === storageModule.StatusCodes.OK) {
			settings = result.$data;
		} else {
			settings = {
				"device-type": "auto"
			};
		}
		// Store last read settings in a global variable to make the accessible from elsewhere
		globals.setGlobalSettings(settings);
		return settings;
	});
};

exports.setGlobalSettings = function(settings) {
	var storage = globals.getStorage();

	var ctx = {
		"$user": "none",
		"$role": "none",
		"$lang": "none"
	};
	return storage.put({
		$context: ctx,
		$collection: "$settings",
		$endpoint: "$local",
		$key: "$global_settings",
		$data: settings
	}).then(function() {
		// Store last stored settings in a global variable to make the accessible from elsewhere
		globals.setGlobalSettings(settings);
		return settings;
	}).fail(function(e) {
		return $.smResolve();
	});
};
/**
 * True if we allow automatic login when we launch the application
 * Should be an option in login dialog
 * Currently returns true only if application runs inside a native wrapper
 */
exports.loginRememberMe = function() {
	return nativeApp.isSageX3WUPApp();
};
/**
 * True to disable automatic login on logout
 * Should be an option in logout dialog
 * Currently returns false only if application runs inside a native wrapper
 */
exports.logoutForgetMe = function() {
	return !nativeApp.isSageX3WUPApp();
};
