// syracuse-tablet/html/js/application/settings.js
"use strict";

/*
 * This module should be used to persist user specific settings
 */

var globals = require('syracuse-tablet/html/js/helpers/globals');
var storageModule = require('syracuse-tablet/html/js/storage/storage');
var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("settings");
var nativeVoiceCommmands = require('syracuse-tablet/html/js/helpers/native/native').getModule("voiceCommands");
var native = require('syracuse-tablet/html/js/helpers/native/native');

var _currentContextUserRole = {};
var _currentContextUser = {};

var _settingsUserRole = null;
var _settingsUser = null;

/*
 *
 */
function _getContext() {
	var ctx = globals.getUserCtx();
	if (!ctx) { // In case not logged in yet!
		return {
			$user: "none",
			$role: "none",
			$lang: "en-US",
		};
	}
	return ctx;
}

/*
 *
 */
function _checkContextUserRole() {
	var ctx = _getContext();
	if (_currentContextUserRole.user != ctx.$user ||
		_currentContextUserRole.role != ctx.$role) {
		_currentContextUserRole.user = ctx.$user;
		_currentContextUserRole.role = ctx.$role;
		_currentContextUserRole.lang = ctx.$lang;
		_settingsUserRole = null;
	}
}

/*
 *
 */
function _readSettingsUserRole() {
	var storage = globals.getStorage();

	var ctx = {
		"$user": _currentContextUserRole.user,
		"$role": _currentContextUserRole.role,
		"$lang": "none"
	};

	return storage.read({
		$context: ctx,
		$collection: "$settings",
		$endpoint: "$local",
		$key: "$user_and_role"
	}).then(function(result) {
		if (result.$status === storageModule.StatusCodes.OK) {
			return result.$data;
		} else {
			return {};
		}
	});
}

/*
 *
 */
function _checkContextUser() {
	var ctx = _getContext();
	if (_currentContextUser.user != ctx.$user ||
		_currentContextUser.role != ctx.$role) {
		_currentContextUser.user = ctx.$user;
		_currentContextUser.role = ctx.$role;
		_currentContextUser.lang = ctx.$lang;
		_settingsUser = null;
	}
}

/*
 *
 */
function _readSettingsUser() {
	if (_settingsUser == null) {
		var storage = globals.getStorage();
		var ctx = {
			"$user": _currentContextUser.user,
			"$role": "none",
			"$lang": "none"
		};
		return storage.read({
			$context: ctx,
			$collection: "$settings",
			$endpoint: "$local",
			$key: "$user"
		}).then(function(result) {
			if (result.$status === storageModule.StatusCodes.OK) {
				_settingsUser = result.$data;
			} else {
				_settingsUser = {};
			}
			log && log("Read user settings...");
			log && log(_settingsUser);
		});
	} else {
		return $.smResolve();
	}
}

/*
 *
 */
function _writeSettingsUser() {
	var storage = globals.getStorage();

	var ctx = {
		"$user": _currentContextUser.user,
		"$role": "none",
		"$lang": "none"
	};

	log && log("Writing user settings...");
	log && log(_settingsUser);
	return storage.put({
		$context: ctx,
		$collection: "$settings",
		$endpoint: "$local",
		$key: "$user",
		$data: _settingsUser
	}).then(function(result) {
		if (result.$status === storageModule.StatusCodes.OK) {
			return result.$data;
		} else {
			return {};
		}
	});
}

/*
 * ======================================================
 * Pages pinned on welcome page
 * ======================================================
 */

/*
 *
 */
exports.getPinnedPages = function() {
	return _readSettingsUser().then(function() {
		_settingsUser.pinnedPages = _settingsUser.pinnedPages || [];
		return _settingsUser.pinnedPages;
	});
};

/*
 *
 */
exports.addPinnedPage = function(page) {
	return exports.getPinnedPages().then(function(pinnedPages) {
		if (!pinnedPages.some(function(p, idx) {
			if (p.id === page.id) {
				pinnedPages[idx] = page; // Overwrite existing
				return true;
			}
		})) {
			pinnedPages.push(page);
		}
		if (nativeVoiceCommmands) {
			nativeVoiceCommmands.addVoiceCommand(nativeVoiceCommmands.createCommandFromPage(page));
		}
		// No return ... then, because can be done async
		_writeSettingsUser();
		return page;
	});
};

/*
 *
 */
exports.isPinnedPage = function(id) {
	return exports.getPinnedPages().then(function(pinnedPages) {
		return pinnedPages.some(function(p, idx) {
			if (p.id === id) {
				return true;
			}
		});
	});
};

/*
 *
 */
exports.removePinnedPage = function(id) {
	return exports.getPinnedPages().then(function(pinnedPages) {
		if (pinnedPages.some(function(p, idx) {
			if (p.id === id) {
				if (nativeVoiceCommmands) {
					nativeVoiceCommmands.removeVoiceCommand(p.title);
				}

				// ID see also welcomeApplication.js if you change here!
				exports.removeTileConfig("pin-" + p.id);

				pinnedPages.splice(idx, 1); // remove and stop
				return true;
			}
		})) {
			// Only write if there was a change
			// No return ... then, because can be done async
			_writeSettingsUser();
		}
	});
};

/*
 * ======================================================
 * Apps hidden on welcome page
 * ======================================================
 */
/*
 *
 */
exports.getHiddenApplications = function() {
	return _readSettingsUser().then(function() {
		_settingsUser.hiddenApplications = _settingsUser.hiddenApplications || [];
		return _settingsUser.hiddenApplications;
	});
};

/*
 *
 */
exports.unhideApplication = function(applicationName) {
	return exports.getHiddenApplications().then(function(apps) {
		if (apps.some(function(app, idx) {
			if (app === applicationName) {
				apps.splice(idx, 1); // remove and stop
				return true;
			}
		})) {
			// Only write if there was a change
			// No return ... then, because can be done async
			_writeSettingsUser();
		}
	});
};

/*
 *
 */
exports.hideApplication = function(applicationName) {
	return exports.getHiddenApplications().then(function(apps) {
		if (!apps.some(function(app, idx) {
			if (app === applicationName) {
				return true;
			}
		})) {
			apps.push(applicationName);
			// No return ... then, because can be done async
			_writeSettingsUser();
		}
	});
};

/*
 * ======================================================
 * Tiles on welcome page
 * ======================================================
 */
exports.getTileConfigs = function() {
	return _readSettingsUser().then(function() {
		_settingsUser.tileConfigs = _settingsUser.tileConfigs || {};
		return _settingsUser.tileConfigs;
	});
};

exports.setTileConfig = function(tile) {
	return exports.getTileConfigs().then(function(configs) {
		configs[tile.tileId] = tile;
		_writeSettingsUser();
	});
};

exports.removeTileConfig = function(id) {
	return exports.getTileConfigs().then(function(configs) {
		if (configs && configs[id]) {
			delete configs[id];
			_writeSettingsUser();
		}
	});
};


/*
 *
 */
exports.getLastUserProfile = function(profile) {
	var ctx = {
		"$user": "none",
		"$role": "none",
		"$lang": "none"
	};
	var storage = globals.getStorage();
	return storage.read({
		$context: ctx,
		$collection: "$settings",
		$endpoint: "$local",
		$key: "$last_user_profile"
	}).then(function(result) {
		if (result.$status === storageModule.StatusCodes.OK) {
			return result.$data;
		} else {
			return null;
		}
	});
};

/*
 *
 */
exports.setLastUserProfile = function(profile) {
	var ctx = {
		"$user": "none",
		"$role": "none",
		"$lang": "none"
	};
	var storage = globals.getStorage();
	return storage.put({
		$context: ctx,
		$collection: "$settings",
		$endpoint: "$local",
		$key: "$last_user_profile",
		$data: profile
	}).then(function(result) {
		if (result.$status === storageModule.StatusCodes.OK) {
			return result.$data;
		} else {
			return {};
		}
	});
};


/*
 * These are settings that are understood to be global for the device
 * This should be things that are common for all users, roles, languages that use the current device
 * E.g. is the device type ("auto", "tablet", "smartphone")
 */
exports.getGlobalSettings = function() {
	var storage = globals.getStorage();

	var ctx = {
		"$user": "none",
		"$role": "none",
		"$lang": "none"
	};

	return storage.read({
		$context: ctx,
		$collection: "$settings",
		$endpoint: "$local",
		$key: "$global_settings"
	}).then(function(result) {
		var settings;
		if (result.$status === storageModule.StatusCodes.OK) {
			settings = result.$data;
		} else {
			settings = {
				"device-type": "auto"
			};
		}
		// Store last read settings in a global variable to make the accessible from elsewhere
		globals.setGlobalSettings(settings);
		return settings;
	});
};

exports.setGlobalSettings = function(settings) {
	var storage = globals.getStorage();

	var ctx = {
		"$user": "none",
		"$role": "none",
		"$lang": "none"
	};
	return storage.put({
		$context: ctx,
		$collection: "$settings",
		$endpoint: "$local",
		$key: "$global_settings",
		$data: settings
	}).then(function() {
		// Store last stored settings in a global variable to make the accessible from elsewhere
		globals.setGlobalSettings(settings);
		return settings;
	}).fail(function(e) {
		return $.smResolve();
	});
};
/*
 * ======================================================
 * Apps hidden on welcome page
 * ======================================================
 */
/**
 * True if we allow automatic login when we launch the application
 * It's an option in login dialog
 */
exports.loginRememberMe = function() {
	return exports.getLoginRememberMe().then(function(val) {
		var val = val || "nativeonly";
		if (val === "always") return true;
		if (val === "never") return false;
		// "nativeonly"
		return native.hasCapability("X3WUPApp");
	});
};
/**
 * True to disable automatic login on logout
 */
exports.logoutForgetMe = function() {
	return exports.getLoginRememberMe().then(function(rememberme) {
		return rememberme === "always" ? false : rememberme === "never" ? true : rememberme === "nativeonly" ? !native.hasCapability("X3WUPApp") : true;
	});
};

exports.getLoginRememberMe = function() {
	return _readSettingsUser().then(function() {
		if (_settingsUser.loginRememberMe == null) {
			return "nativeonly";
		}
		return _settingsUser.loginRememberMe;
	});
};
exports.setLoginRememberMe = function(value) {
	return _readSettingsUser().then(function(apps) {
		if (_settingsUser.loginRememberMe !== value) {
			_settingsUser.loginRememberMe = value;
			_writeSettingsUser();
		}
	});
};
