// syracuse-tablet/html/js/application/settings.js
"use strict";

/*
 * This module should be used to persist user specific settings
 */

var globals = require('syracuse-tablet/html/js/helpers/globals');
var storageModule = require('syracuse-tablet/html/js/storage/storage');

var _currentContextUserRole = {};
var _currentContextUser = {};

var _settingsUserRole = null;
var _settingsUser = null;

function _checkContextUserRole() {
	var ctx = globals.getUserCtx();
	if (_currentContextUserRole.user != ctx.$user ||
		_currentContextUserRole.role != ctx.$role) {
		_currentContextUserRole.user = ctx.$user;
		_currentContextUserRole.role = ctx.$role;
		_currentContextUserRole.lang = ctx.$lang;
		_settingsUserRole = null;
	}
}

function _readSettingsUserRole() {
	var storage = globals.getStorage();

	var ctx = {
		"$user": _currentContextUserRole.user,
		"$role": _currentContextUserRole.role,
		"$lang": "none"
	};

	return storage.read({
		$context: ctx,
		$collection: "$settings",
		$endpoint: "$local",
		$key: "$user_and_role"
	}).then(function(result) {
		if (result.$status === storageModule.StatusCodes.OK) {
			return result.$data;
		} else {
			return {};
		}
	});
}

function _checkContextUser() {
	var ctx = globals.getUserCtx();
	if (_currentContextUser.user != ctx.$user ||
		_currentContextUser.role != ctx.$role) {
		_currentContextUser.user = ctx.$user;
		_currentContextUser.role = ctx.$role;
		_currentContextUser.lang = ctx.$lang;
		_settingsUser = null;
	}
}

function _readSettingsUser() {
	var storage = globals.getStorage();

	var ctx = {
		"$user": _currentContextUser.user,
		"$role": "none",
		"$lang": "none"
	};

	return storage.read({
		$context: ctx,
		$collection: "$settings",
		$endpoint: "$local",
		$key: "$user"
	}).then(function(result) {
		if (result.$status === storageModule.StatusCodes.OK) {
			_settingsUser = result.$data;
		} else {
			_settingsUser = {};
		}
	});
}

function _writeSettingsUser() {
	var storage = globals.getStorage();

	var ctx = {
		"$user": _currentContextUser.user,
		"$role": "none",
		"$lang": "none"
	};

	return storage.put({
		$context: ctx,
		$collection: "$settings",
		$endpoint: "$local",
		$key: "$user",
		$data: _settingsUser
	}).then(function(result) {
		if (result.$status === storageModule.StatusCodes.OK) {
			return result.$data;
		} else {
			return {};
		}
	});
}

/*
 *
 */
exports.getPinnedPages = function() {
	_checkContextUser();
	var read;
	if (_settingsUser == null) {
		read = _readSettingsUser();
	} else {
		read = $.smResolve();
	}
	return read.then(function() {
		_settingsUser.pinnedPages = _settingsUser.pinnedPages || [];
		return _settingsUser.pinnedPages;
	});
};

/*
 *
 */
exports.addPinnedPage = function(page) {
	return exports.getPinnedPages().then(function(pages) {
		if (!pages.some(function(p, idx) {
			if (p.id === page.id) {
				pages[idx] = page; // Overwrite existing
				return true;
			}
		})) {
			pages.push(page);
		}
		// No return ... then, because can be done async
		_writeSettingsUser();
		return page;
	});
};

exports.isPinnedPage = function(id) {
	return exports.getPinnedPages().then(function(pages) {
		return pages.some(function(p, idx) {
			if (p.id === id) {
				return true;
			}
		});
	});
};

/*
 *
 */
exports.removePinnedPage = function(id) {

};
