// syracuse-ui/lib/extend/extendField.js
"use strict";
var _helpers = require('syracuse-core').helpers;
var Field = require('syracuse-ui/lib/field/field').Field;

function ExtendField() {}

exports.ExtendField = _helpers.defineClass(ExtendField, Field, {
	load: function(initData) {
		this.$item.$isTitleHidden = true;
		this.$item.$isTopLabelAlignment = true;
		if (this.api.stylesheets) {
			for (var ii = 0, jj = this.api.stylesheets.length; ii < jj; ii++) {
				syra_site.loadStyleSheet(this.api.stylesheets[ii], true);
			}
		}

		if (this.$field.$type.indexOf("x-array") >= 0) {
			syra_article.add(this);
			this.domItem = this._dataValue = this.layoutSlot;
			this.render(this.layoutSlot);
			syra_article.endDraw(this, initData);
		} else {
			Field.prototype.load.call(this);
		}

	},
	addValueSlot: function() {
		var css = this._$cssField + (this.$isEditMode ? "-value-edit" : "-value-read");
		this.domItem.className += " " + this._$cssField;
		this._core = this.domItem;
		this._dataValue = syra_dom.div("", this.domItem);
	},
	render: function(slot) {
		var self = this;

		self.extendWidget = self.api.create({
			id: self.id,
			field: self,
			prototype: self.$prototype || self.$field, //pas article
			div: self._dataValue,
			design: self.$item,
			facet: self.page.$facet,
			editable: self.$isEditMode,
			onFieldInputEvent: function(event, shortcuts) {
				return self.onFieldInputEvent(event, shortcuts);
			},
			invalidate: function(error) {
				syra_fields.invalidate(self, error);
			},
			setDirty: function(dirty) {
				self.setDirty(dirty !== false);
				if (self.$isEditMode) {
					self.page.applyChange({
						$actions: {
							$save: {
								$isDisabled: false,
							}
						}
					});
				}
			}
		});
		if (self.extendWidget.resize) {
			if (self.api.isResizable) {
				$(self.domItem).resizable({
					handles: "n,s",
					resize: function(event, ui) {
						self.extendWidget.resize(ui);
					}
				});
			}
			self.$item.$isAutoSize = true;
		}
	},
	validate: function(errors, value) {
		if (this.extendWidget && this.extendWidget.validate) {
			this.extendWidget.validate(errors, value);
		}
	},
	setState: function(state) {
		syra_fields.setState(this, state);
		if (this.extendWidget && this.extendWidget.onStateChanged) {
			this.extendWidget.onStateChanged(state);
		}
	},
	setDirty: function(isDirty) {
		Field.prototype.setDirty.call(this, isDirty);
		this.extendWidget._isDirty = isDirty;
		if (this.extendWidget && this.extendWidget.onDirtyChanged) {
			this.extendWidget.onDirtyChanged(isDirty);
		}
	},
	isDirty: function() {
		if (this.extendWidget && this.extendWidget.isDirty) {
			return this.extendWidget.isDirty();
		}
		return this.extendWidget._isDirty;
	},
	getValue: function() {
		if (this.extendWidget && this.extendWidget.getValue) {
			return this.extendWidget.getValue();
		} else {
			return this.currentValue;
		}
	},
	setValue: function(value, metaData, parentDataRecord) {
		this.currentValue = value;
		if (this.extendWidget && this.extendWidget.setValue) {
			this.extendWidget.setValue(value, metaData, parentDataRecord);
		}
		this.applyMetaData(metaData);
	},

	applyShortCut: function(shortcuts, event) {
		if (this.extendWidget && this.extendWidget.applyShortCut) {
			return this.extendWidget.applyShortCut(shortcuts, event);
		}
		return false;
	},
	setInputFocus: function(select) {
		if (this.extendWidget && this.extendWidget.setInputFocus) {
			this.extendWidget.setInputFocus(select);
		} else {
			Field.prototype.setInputFocus.call(this, select);
		}
	},
	getInputValue: function() {
		if (this.extendWidget && this.extendWidget.getInputValue) {
			return this.extendWidget.getInputValue();
		}
		return Field.prototype.getInputValue.call(this);
	},
	getCaretPosition: function() {
		if (this.extendWidget && this.extendWidget.getCaretPosition) {
			return this.extendWidget.getCaretPosition();
		}
		return -1;
	},
	applyDesignMeta: function(metadata, designing) {
		if (this.extendWidget && this.extendWidget.applyDesignMeta)
			this.extendWidget.applyDesignMeta(metadata, designing);
	},
	resizeItem: function() {
		if (this.extendWidget && this.extendWidget.resize) {
			if (syra_dom.isVisible(this._dataValue)) {
				this.extendWidget.resize();
			}
		}
	},
	dispose: function() {
		this.extendWidget && this.extendWidget.dispose && this.extendWidget.dispose();
		syra_site.disposeObject(this.extendWidget);
		if (this.isArticle) {
			syra_article.dispose(this);
		} else {
			Field.prototype.dispose.call(this);
		}
	}
});
