// syracuse-x3/lib/contract-extension.ts
"use strict";

module.exports = {
	application: "syracuse",
	extends: "collaboration",
	resources: require('./strings').module,
	entities: {
		endPoint: require('./entities/endpoint-extension'),

		boServer: require('./entities/endpoint-extension'),
		// Business Objects
		boServer: require('./entities/bo/boServer').entity,
		boProfile: require('./entities/bo/boProfile').entity,

		cvgSession: require('./entities/cvgSession').entity,
		cvgReuseClient: require('./entities/cvgReuseClient').entity,
		cvgRecord: require('./entities/cvgRecord').entity,

		lookupX3Folder: require('./entities/lookupX3Folder').entity,
		roleBoProfile: require('./entities/user/roleBoProfile').entity,
		sessionInfo: require('./entities/sessionInfo').entity,
		update: require('./entities/update').entity,
		v6DocumentProxy: require('./entities/storage/v6DocumentProxy').entity,

		x3ClassAction: require('./entities/introspection/x3Task/x3ClassAction').entity,
		x3ClassActionParam: require('./entities/introspection/x3Task/x3ClassActionParam').entity,
		x3ClassActionDateParam: require('./entities/introspection/x3Task/x3ClassActionDateParam').entity,
		x3RightsCache: require('./entities/user/x3RightsCache').entity,
		x3SessionConfig: require('./entities/x3SessionConfig').entity,
		x3Task: require('./entities/introspection/x3Task/x3Task').entity,
		x3server: require('./entities/x3server').entity,
		x3solution: require('./entities/x3solution').entity,
		x3UserImport: require('./entities/user/x3UserImport').entity,
		patchGenerator: require('./entities/patchGenerator').entity,
	},
	representations: {
		cvgSession: require('./representations/cvgSession').representation,

		sessionInfo: require('./representations/sessionInfo').representation,
		update: require('./representations/update').representation,

		x3SessionConfig: require('./representations/x3SessionConfig').representation,
		x3Task: require('./representations/introspection/x3Task').representation,
		x3UserImport: require('./representations/x3UserImport').representation,

	},
	dbMeta: {
		initScript: [],
		updateScript: []
	}
};
