// syracuse-tablet/test/js/widgets/resources.js
"use strict";

/*
 * LAYOUT RESOURCES
 */
// NEW V1
var $articleSample1 = {
	$layoutType: "row",
	$widthMd: "6,6",
	$items: [{
		$title: "Dashboard",
		$layoutType: "row",
		$widthMd: "8,4",
		$items: [{
			$bind: "tile 1"
		}, {
			$bind: "tile 2"
		}]
	}, {
		$title: "Apps",
		$layoutType: "stack",
		$items: [{
			$layoutType: "row",
			$widthMd: "6,6",
			$items: [{
				$bind: "tile 3"
			}, {
				$bind: "tile 4"
			}]
		}, {
			$layoutType: "row",
			$widthMd: "6,6",
			$items: [{
				$bind: "tile 5"
			}, {
				$bind: "tile 6"
			}]
		}, {
			$layoutType: "row",
			$widthMd: "6,6",
			$items: [{
				$bind: "tile 7"
			}, {
				$bind: "tile 8"
			}]
		}]

	}]

};

// NEW V2

var $articleSample2 = {
	$layoutType: "row",
	$widthMd: "6,6",
	$items: [{
		$title: "Dashboard",
		$layoutType: "row",
		$widthMd: "8,4",
		$items: [{
			$bind: "tile 1"
		}, {
			$bind: "tile 2"
		}]

	}, {
		$title: "Apps",
		$layoutType: "row",
		$widthMd: "6,6",
		$items: [{
			$layoutType: "stack",
			$items: [{
				$bind: "tile 3"
			}, {
				$bind: "tile 5"
			}, {
				$bind: "tile 7"
			}]
		}, {
			$layoutType: "stack",
			$items: [{
				$bind: "tile 4"
			}, {
				$bind: "tile 6"
			}, {
				$bind: "tile 8"
			}]
		}]

	}]
};

var $articleSample3 = {
	$layoutType: "row",
	$widthMd: "8,4",
	$items: [{
		$title: "Details",
		$layoutType: "row",
		$widthMd: "4,8",
		$items: [{
			$bind: "tile 1"
		}, {
			$layoutType: "stack",
			$items: [{
				$bind: "tile 2"
			}, {
				$layoutType: "row",
				$widthMd: "6,6",
				$items: [{
					$bind: "tile 3"
				}, {
					$bind: "tile 4"
				}]
			}, {
				$layoutType: "row",
				$widthMd: "6,6",
				$items: [{
					$bind: "tile 5"
				}, {
					$bind: "tile 6"
				}]
			}]
		}]
	}, {
		$title: "Statistics",
		$layoutType: "row",
		$widthMd: "6,6",
		$items: [{
			$layoutType: "stack",
			$items: [{
				$bind: "tile 7"
			}, {
				$bind: "tile 9"
			}, {
				$bind: "tile 11"
			}]
		}, {
			$layoutType: "stack",
			$items: [{
				$bind: "tile 8"
			}, {
				$bind: "tile 10"
			}, {
				$bind: "tile 12"
			}]
		}]
	}]
};

var $articleSample4 = {
	$widthMd: "4,8",
	$layoutType: "row",
	$items: [{
		$title: "Tuesday, May 13",
		$layoutType: "stack",
		$items: [{
			$layoutType: "tabs",
			$items: [{
				$title: "Today",
				$layoutType: "stack",
				$items: [{
					$bind: "tile 1. Appointments (today)"
				}, {
					$bind: "tile 2. To do (today)"
				}]
			}, {
				$title: "Tomorrow",
				$layoutType: "stack",
				$items: [{
					$bind: "tile 1. Appointments (tomorrow)"
				}, {
					$bind: "tile 2. To do (tomorrow)"
				}]
			}]
		}]
	}, {
		$title: "Welcome",
		$layoutType: "row",
		$widthMd: "6,6",
		$items: [{
			$layoutType: "stack",
			$items: [{
				$layoutType: "row",
				$widthMd: "6,6",
				$items: [{
					$bind: "tile 3. Short dated quotes"
				}, {
					$bind: "tile 4. Products to push"
				}]
			}, {
				$layoutType: "row",
				$widthMd: "6,6",
				$items: [{
					$bind: "tile 6. Payment reminder"
				}, {
					$bind: "tile 7. Search price"
				}]
			}, {
				$bind: "tile 9. WIP"
			}]
		}, {
			$layoutType: "stack",
			$items: [{
				$bind: "tile 5. My goals"
			}, {
				$bind: "tile 8. Turn over N/N-"
			}, {
				$bind: "tile 10. Inventory"
			}]
		}]
	}]
};

var $articleSample5 = {
	$layoutType: "stack",
	$items: [{
		$layoutType: "row",
		$widthMd: "8,2,2",
		$items: [{
			$bind: "TILE 1. Appointments"
		}, {
			$bind: "TILE 2. TO DO"
		}, {
			$bind: "TILE 3. Memo"
		}]
	}, {
		$layoutType: "row",
		$widthMd: "4,8",
		$items: [{
			$bind: "TILE 4. Call history"
		}, {
			$layoutType: "stack",
			$items: [{
				$layoutType: "row",
				$widthMd: "4,4,4",
				$items: [{
					$bind: "TILE 5. WIP Order"
				}, {
					$bind: "TILE 6. WIP Order"
				}, {
					$bind: "TILE 8. WIP Delivery"
				}]
			}, {
				$layoutType: "row",
				$widthMd: "8,4",
				$items: [{
					$bind: "TILE 7. Order book"
				}, {
					$bind: "TILE 9. Payment reminder"
				}]
			}]

		}]
	}]
};

exports.$articleSample1 = $articleSample1;
exports.$articleSample2 = $articleSample2;
exports.$articleSample3 = $articleSample3;
exports.$articleSample4 = $articleSample4;
exports.$articleSample5 = $articleSample5;

/*
 * FIELD RESOURCES
 */
/**
 * testCase structure - title : Test case title (widget type) - editContext :
 * template sample context for edit mode - detailContext : template sample
 * context for detail mode - type : application type
 */
var fieldTestCases = [{
	title: "X-STRING",
	editContext: {
		$title: "Username",
		$isEditMode: true,
		value: "Tyrion Lannister"
	},
	detailContext: {
		$title: "Username",
		value: "Tyrion Lannister"
	},
	type: "application/x-string"
}, {
	title: "X-STRING:PHONE",
	editContext: {
		isPhone: true,
		$title: "Phone",
		$isEditMode: true,
		value: "08.10.30.30.30"
	},
	detailContext: {
		isPhone: true,
		$title: "Phone",
		value: "08.10.30.30.30"
	},
	type: "application/x-string"
}, {
	title: "X-STRING:EMAIL",
	editContext: {
		$title: "Email",
		$isEditMode: true,
		isEmail: true,
		value: "tyrion.lannister@kingslanding.com"
	},
	detailContext: {
		$title: "Email",
		isEmail: true,
		value: "tyrion.lannister@kingslanding.com"
	},
	type: "application/x-string"
}, {
	title: "X-INTEGER/DECIMAL",
	editContext: {
		$title: "Age",
		$isEditMode: true,
		value: "35"
	},
	detailContext: {
		$title: "Age",
		value: "35"
	},
	type: "application/x-integer"
}, {
	title: "X-BOOLEAN",
	editContext: {
		$title: "Always pays his debt ?",
		$isEditMode: true,
		value: true
	},
	detailContext: {
		$title: "Always pays his debt ?",
		value: true
	},
	type: "application/x-boolean"
}, {
	title: "X-CHOICE:RADIO",
	editContext: {
		$title: "Your family favorite :",
		isRadio: true,
		$isEditMode: true,
		choiceItem: [{
			itemValue: "Cersei"
		}, {
			itemValue: "Jaime",
			isSelected: true
		}, {
			itemValue: "Robert"
		}, {
			itemValue: "Tywin"
		}, {
			itemValue: "Joanna"
		}]
	},
	detailContext: {
		$title: "Your family favorite :",
		value: "Jaime"
	},
	type: "application/x-choice"
}, {
	title: "X-CHOICE:COMBO",
	editContext: {
		$title: "Your family favorite :",
		$isEditMode: true,
		choiceItem: [{
			itemValue: "Cersei"
		}, {
			itemValue: "Jaime",
			isSelected: true
		}, {
			itemValue: "Robert"
		}, {
			itemValue: "Tywin"
		}, {
			itemValue: "Joanna"
		}],
		value: "Jaime"
	},
	detailContext: {
		$title: "Your family favorite :",
		value: "Jaime"
	},
	type: "application/x-choice"
}, {
	title: "PROGRESS",
	type: "application/x-progress",
	editContext: {
		$title: "How close to beeing killed by the author, George R. R. Martin ?",
		value: 99,
		valueMin: 0,
		valueMax: 100,
		valueNow: 99
	},
	detailContext: {
		$title: "How close to beeing killed by the author, George R. R. Martin ?",
		value: 99,
		valueMin: 0,
		valueMax: 100,
		valueNow: 99
	}
}];
exports.fieldTestCases = fieldTestCases;
