// syracuse-tablet/test/js/widgets/resources.js
"use strict";

// NEW V1
var $articleSample1 = {
	$layoutType: "row",
	$widthsDesktop: "6,6",
	$items: [{
		$title: "Dashboard",
		$layoutType: "row",
		$widthsDesktop: "8,4",
		$items: [{
			$bind: "tile 1"
		}, {
			$bind: "tile 2"
		}]
	}, {
		$title: "Apps",
		$layoutType: "stack",
		$items: [{
			$layoutType: "row",
			$widthsDesktop: "6,6",
			$items: [{
				$bind: "tile 3"
			}, {
				$bind: "tile 4"
			}]
		}, {
			$layoutType: "row",
			$widthsDesktop: "6,6",
			$items: [{
				$bind: "tile 5"
			}, {
				$bind: "tile 6"
			}]
		}, {
			$layoutType: "row",
			$widthsDesktop: "6,6",
			$items: [{
				$bind: "tile 7"
			}, {
				$bind: "tile 8"
			}]
		}]

	}]

};
exports.$articleSample1 = $articleSample1;

// NEW V2

var $articleSample2 = {
	$layoutType: "row",
	$widthsDesktop: "6,6",
	$items: [{
		$title: "Dashboard",
		$layoutType: "row",
		$widthsDesktop: "8,4",
		$items: [{
			$bind: "tile 1"
		}, {
			$bind: "tile 2"
		}]

	}, {
		$title: "Apps",
		$layoutType: "row",
		$widthsDesktop: "6,6",
		$items: [{
			$layoutType: "stack",
			$items: [{
				$bind: "tile 3"
			}, {
				$bind: "tile 5"
			}, {
				$bind: "tile 7"
			}]
		}, {
			$layoutType: "stack",
			$items: [{
				$bind: "tile 4"
			}, {
				$bind: "tile 6"
			}, {
				$bind: "tile 8"
			}]
		}]

	}]
};
exports.$articleSample2 = $articleSample2;

var $articleSample3 = {
	$layoutType: "row",
	$widthsDesktop: "8,4",
	$items: [{
		$title: "Details",
		$layoutType: "row",
		$widthsDesktop: "4,8",
		$items: [{
			$bind: "tile 1"
		}, {
			$layoutType: "stack",
			$items: [{
				$bind: "tile 2"
			}, {
				$layoutType: "row",
				$widthsDesktop: "6,6",
				$items: [{
					$bind: "tile 3"
				}, {
					$bind: "tile 4"
				}]
			}, {
				$layoutType: "row",
				$widthsDesktop: "6,6",
				$items: [{
					$bind: "tile 5"
				}, {
					$bind: "tile 6"
				}]
			}]
		}]
	}, {
		$title: "Statistics",
		$layoutType: "row",
		$widthsDesktop: "6,6",
		$items: [{
			$layoutType: "stack",
			$items: [{
				$bind: "tile 7"
			}, {
				$bind: "tile 9"
			}, {
				$bind: "tile 11"
			}]
		}, {
			$layoutType: "stack",
			$items: [{
				$bind: "tile 8"
			}, {
				$bind: "tile 10"
			}, {
				$bind: "tile 12"
			}]
		}]
	}]
};
exports.$articleSample3 = $articleSample3;

var $articleSample4 = {
	$widthsDesktop: "4,8",
	$layoutType: "row",
	$items: [{
		$title: "Tuesday, May 13",
		$layoutType: "stack",
		$items: [{
			$layoutType: "tabs",
			$items: [{
				$title: "Today",
				$layoutType: "stack",
				$items: [{
					$bind: "tile 1. Appointments (today)"
				}, {
					$bind: "tile 2. To do (today)"
				}]
			}, {
				$title: "Tomorrow",
				$layoutType: "stack",
				$items: [{
					$bind: "tile 1. Appointments (tomorrow)"
				}, {
					$bind: "tile 2. To do (tomorrow)"
				}]
			}]
		}]
	}, {
		$title: "Welcome",
		$layoutType: "row",
		$widthsDesktop: "6,6",
		$items: [{
			$layoutType: "stack",
			$items: [{
				$layoutType: "row",
				$widthsDesktop: "6,6",
				$items: [{
					$bind: "tile 3. Short dated quotes"
				}, {
					$bind: "tile 4. Products to push"
				}]
			}, {
				$layoutType: "row",
				$widthsDesktop: "6,6",
				$items: [{
					$bind: "tile 6. Payment reminder"
				}, {
					$bind: "tile 7. Search price"
				}]
			}, {
				$bind: "tile 9. WIP"
			}]
		}, {
			$layoutType: "stack",
			$items: [{
				$bind: "tile 5. My goals"
			}, {
				$bind: "tile 8. Turn over N/N-"
			}, {
				$bind: "tile 10. Inventory"
			}]
		}]
	}]
};
exports.$articleSample4 = $articleSample4;
