// syracuse-tablet/test/js/widgets/resources.js
"use strict";

// NEW V1
var $articleSample1 = {
	$layout: {
		$items: [{
			$layoutType: "row",
			$widthsDesktop: "6,6",
			$items: [{
				$category: "section",
				$title: "Dashboard",
				$layout: {
					$layoutType: "row",
					$widthsDesktop: "8,4",
					$items: [{
						$bind: "tile 1"
					}, {
						$bind: "tile 2"
					}]
				}
			}, {
				$category: "section",
				$title: "Apps",
				$layout: {
					$layoutType: "stack",
					$items: [{
						$layoutType: "row",
						$widthsDesktop: "6,6",
						$items: [{
							$bind: "tile 3"
						}, {
							$bind: "tile 4"
						}]
					}, {
						$layoutType: "row",
						$widthsDesktop: "6,6",
						$items: [{
							$bind: "tile 5"
						}, {
							$bind: "tile 6"
						}]
					}, {
						$layoutType: "row",
						$widthsDesktop: "6,6",
						$items: [{
							$bind: "tile 7"
						}, {
							$bind: "tile 8"
						}]
					}]
				}
			}]

		}]
	}
};
exports.$articleSample1 = $articleSample1;

// NEW V2

var $articleSample2 = {
	$layout: {
		$items: [{
			$layoutType: "row",
			$widthsDesktop: "6,6",
			$items: [{
				$category: "section",
				$title: "Dashboard",
				$layout: {
					$layoutType: "row",
					$widthsDesktop: "8,4",
					$items: [{
						$bind: "tile 1"
					}, {
						$bind: "tile 2"
					}]
				}
			}, {
				$category: "section",
				$title: "Apps",
				$layout: {
					$layoutType: "row",
					$widthsDesktop: "6,6",
					$items: [{
						$layoutType: "stack",
						$items: [{
							$bind: "tile 3"
						}, {
							$bind: "tile 5"
						}, {
							$bind: "tile 7"
						}]
					}, {
						$layoutType: "stack",
						$items: [{
							$bind: "tile 4"
						}, {
							$bind: "tile 6"
						}, {
							$bind: "tile 8"
						}]
					}]
				}
			}]
		}]
	}
};
exports.$articleSample2 = $articleSample2;

var $articleSample3 = {
	$layout: {
		$items: [{
			$layoutType: "row",
			$widthsDesktop: "8,4",
			$items: [{
				$category: "section",
				$title: "Details",
				$layout: {
					$layoutType: "row",
					$widthsDesktop: "4,8",
					$items: [{
						$bind: "tile 1"
					}, {
						$layoutType: "stack",
						$items: [{
							$bind: "tile 2"
						}, {
							$layoutType: "row",
							$widthsDesktop: "6,6",
							$items: [{
								$bind: "tile 3"
							}, {
								$bind: "tile 4"
							}]
						}, {
							$layoutType: "row",
							$widthsDesktop: "6,6",
							$items: [{
								$bind: "tile 5"
							}, {
								$bind: "tile 6"
							}]
						}]
					}]
				}
			}, {
				$category: "section",
				$title: "Statistics",
				$layout: {
					$layoutType: "row",
					$widthsDesktop: "6,6",
					$items: [{
						$layoutType: "stack",
						$items: [{
							$bind: "tile 7"
						}, {
							$bind: "tile 9"
						}, {
							$bind: "tile 11"
						}]
					}, {
						$layoutType: "stack",
						$items: [{
							$bind: "tile 7"
						}, {
							$bind: "tile 9"
						}, {
							$bind: "tile 11"
						}]
					}]
				}
			}]
		}]
	}
};
exports.$articleSample3 = $articleSample3;
