// syracuse-ui/lib/page/landing/designer/landingPageDesigner.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Designer = require("syracuse-ui/lib/authoring/designer").Designer;
var UISwitchTool = require("syracuse-ui/lib/authoring/tools/uiSwitchTool").UISwitchTool;
var ItemTool = require("syracuse-ui/lib/authoring/tools/itemTool").ItemTool;

function LandingPageDesigner() {}

exports.LandingPageDesigner = helpers.defineClass(LandingPageDesigner, Designer, {
	loadBox: function() {
		this.$prototype = {
			$properties: {
				editTab: {
					$type: "application/x-string",
					$title: syra_local.ldpDesignTitle,
					$capabilities: "localize",
					$links: {
						$localize: {
							$title: syra_local.ldpDesignTranslation
						}
					}
				}
			}
		};
		Designer.prototype.loadBox.call(this);
	},

	drawBox: function() {
		this.layoutSlot = this.domItem = document.createElement("aside");
		this.domItem.syraItem = this.domItem.syraarticle = this.id;
		this.layoutSlot.className = "s-aw-lp-designer";
		this.designedItem = this.designedArticle;
		this.designedArticle.header.appendChild(this.layoutSlot);
		this.isVisible = false;
		this.toggleDesigner(false);
		this.addClosePicker(this.layoutSlot);

		this.body = document.createElement("div");
		this.body.className = "s-aw-lp-designer-body";
		this.layoutSlot.appendChild(this.body);
		(this.itemTool = new ItemTool()).load(this);

		this.loadHistoryTool();

		var div = document.createElement("div");
		div.className = "s-aw-lp-designer-tab-edit";
		this.body.appendChild(div);
		this.editTitleField = this.page.loadNewItem(div, {
			$bind: "editTab",
			$category: "field",
			$css: "s-aw-prop",
			$useLocalizePicker: true,
			$isEditMode: true,
			$isTopLabelAlignment: false
		});

		this.sectionModel = {
			body: this.body,
			dataStore: {
				modelSlot: document.createElement("div")
			}
		};
		this.sectionModel.dataStore.modelSlot.className = "s-aw-lp-designer-templates";
		this.sectionModel.body.appendChild(this.sectionModel.dataStore.modelSlot);
		(this.uiSwitchTool = new UISwitchTool()).load(this);

		this.loadTemplateTools();


		var link = document.createElement("a");
		link.className = "s-aw-exclude-me";
		link.syraOnClick = "onExcludeTabClick";
		link.textContent = link.title = syra_local.ldpDelete;
		var div = document.createElement("div");
		div.className = "s-aw-lp-designer-tab-delete";
		div.appendChild(link);
		this.body.appendChild(div);

		this.uiSwitchTool.toggleUIDesign(this.designedArticle, true);
	},
	notifyDataChange: function(field, value) {
		if (this.editTitleField == field && this.editTitleField.localizedSection) {
			syra_site.localizer.onSectionTitleEditorChange(field, value);
		}
		Designer.prototype.notifyDataChange.call(this, field, value);
	},
	onEndHistoryChangeStep: function(itemToSelect) {
		this.uiSwitchTool.toggleUIDesign(this.designedArticle, true);
	},
	endArticleUpdate: function(selectedItem, isStructureUpdated, saveUpdate) {
		this.designedArticle.ensurePageVisibility();
		this.uiSwitchTool.toggleUIDesign(this.designedArticle, true);
		this.historyTool.notifyUpdate();
		if (saveUpdate !== false) {
			this.saveDesign();
		}
		this.onTabClik();
	},
	saveDesign: function() {
		this.designedArticle.landingPageMaster.saveLandingDesign(this.designedArticle);
	},
	onExcludeTabClick: function() {
		this.designedArticle.deleteOpenedTab();
	},
	onTabClik: function(tabOpened) {
		if (this.isVisible) {
			tabOpened = tabOpened || this.designedArticle.layoutContent.getOpenedTab();
			if (tabOpened) {
				syra_site.localizer.setSectionTitleEditor(this.editTitleField, tabOpened);
				this.templatesTool.addLandingPageModels(this.sectionModel, tabOpened);
			} else {
				this.toggleDesigner(false);
			}
		}
	},
	selectItem: function() {

	},
	closeDesigner: function() {
		this.toggleDesigner(false);
	},
	toggleDesigner: function(show) {
		this.isVisible = show === undefined ? !this.isVisible : show;
		syra_site.dom.disableItem(this.designedArticle.designPicker, this.isVisible);
		this.layoutSlot.style.display = this.isVisible ? "" : "none";
		this.onTabClik();
		this.ensurePageVisibility();
	},
	dispose: function() {
		if (this.itemTool) {
			this.itemTool.dispose();
		}
		if (this.domItem) {
			syra_site.dom.removeChild(this.domItem);
		}
		if (this.editTitleField) {
			this.editTitleField.localizedSection = null;
		}
		this.editTitleField = this.itemTool = this.sectionModel = null;
		Designer.prototype.dispose.call(this);
	}
});
