// syracuse-ui/lib/page/landing/designer/landingPageDesigner.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Designer = require("syracuse-ui/lib/authoring/designer").Designer;
var UISwitchTool = require("syracuse-ui/lib/authoring/tools/uiSwitchTool").UISwitchTool;
var ItemTool = require("syracuse-ui/lib/authoring/tools/itemTool").ItemTool;

function LandingPageDesigner() {}

exports.LandingPageDesigner = helpers.defineClass(LandingPageDesigner, Designer, {
	loadBox: function() {
		this.$prototype = {
			$properties: {
				editTab: {
					$type: "application/x-string",
					$title: this.localize.ldpDesignTitle,
					$capabilities: "localize",
					$links: {
						$localize: {
							$title: this.localize.ldpDesignTranslation
						}
					}
				}
			}
		};
		Designer.prototype.loadBox.call(this);
	},

	drawBox: function() {
		this.layoutSlot = this.domItem = document.createElement("aside");
		this.layoutSlot.className = "s-aw-lp-designer";
		this.designedItem = this.designedArticle;
		this.designedArticle.header.appendChild(this.layoutSlot);
		this.isVisible = false;
		this.toggleDesigner(false);
		this.addClosePicker(this.layoutSlot);

		this.body = document.createElement("div");
		this.body.className = "s-aw-lp-designer-body";
		this.layoutSlot.appendChild(this.body);
		(this.itemTool = new ItemTool()).load(this);

		this.loadHistoryTool();

		var div = document.createElement("div");
		div.className = "s-aw-lp-designer-tab-edit";
		this.body.appendChild(div);
		this.editTitleField = this.page.loadNewItem(div, {
			$bind: "editTab",
			$category: "field",
			$css: "s-aw-prop",
			$useLocalizePicker: true,
			$isEditMode: true,
			$isTopLabelAlignment: false
		});

		this.sectionModel = {
			body: this.body,
			dataStore: {
				modelSlot: document.createElement("div")
			}
		};
		this.sectionModel.dataStore.modelSlot.className = "s-aw-lp-designer-templates";
		this.sectionModel.body.appendChild(this.sectionModel.dataStore.modelSlot);
		(this.uiSwitchTool = new UISwitchTool()).load(this);

		this.loadTemplateTools();


		this.excludeMeLink = document.createElement("a");
		this.excludeMeLink.className = "s-aw-exclude-me";
		this.excludeMeLink.textContent = this.excludeMeLink.title = this.localize.ldpDelete;
		this.excludeMeLink.setAttribute("data-s-picker", "$excludeMe");
		var div = document.createElement("div");
		div.className = "s-aw-lp-designer-tab-delete";
		div.appendChild(this.excludeMeLink);
		this.body.appendChild(div);

		this.uiSwitchTool.toggleUIDesign(this.designedArticle, true);
	},
	notifyDataChange: function(field, value) {
		if (this.editTitleField == field && this.editTitleField.localizedSection) {
			document.site.localizer.onSectionTitleEditorChange(field, value);
		}
		Designer.prototype.notifyDataChange.call(this, field, value);
	},
	onEndHistoryChangeStep: function(itemToSelect) {
		this.uiSwitchTool.toggleUIDesign(this.designedArticle, true);
	},
	endArticleUpdate: function(selectedItem, isStructureUpdated, saveUpdate) {
		this.designedArticle.ensurePageVisibility();
		this.uiSwitchTool.toggleUIDesign(this.designedArticle, true);
		this.historyTool.notifyUpdate();
		if (saveUpdate !== false) {
			this.saveDesign();
		}
		this.onTabClik();
	},
	saveDesign: function() {
		this.designedArticle.landingPageMaster.saveLandingDesign(this.designedArticle);
	},
	onClickPicker: function(picker) {
		switch (picker.getAttribute("data-s-picker")) {
			case "$excludeMe":
				this.designedArticle.deleteOpenedTab();
				return false;
		}
		if (!Designer.prototype.onClickPicker.call(this, picker)) {
			return false;
		}
		if (this.uiSwitchTool && this.uiSwitchTool.onClickPicker(picker)) {
			return false;
		}
		return true;
	},
	onTabClik: function(tabOpened) {
		if (this.isVisible) {
			tabOpened = tabOpened || this.designedArticle.layoutContent.getOpenedTab();
			if (tabOpened) {
				document.site.localizer.setSectionTitleEditor(this.editTitleField, tabOpened);
				this.templatesTool.addLandingPageModels(this.sectionModel, tabOpened);
			} else {
				this.toggleDesigner(false);
			}
		}
	},
	selectItem: function() {

	},
	closeDesigner: function() {
		this.toggleDesigner(false);
	},
	toggleDesigner: function(show) {
		this.isVisible = show === undefined ? !this.isVisible : show;
		document.site.dom.disableItem(this.designedArticle.designPicker, this.isVisible);
		this.layoutSlot.style.display = this.isVisible ? "" : "none";
		this.onTabClik();
		this.ensurePageVisibility();
	},
	onTemplateChoiceClick: function(picker, designedItem, awLayout) {
		designedItem = this.designedArticle.layoutContent.getOpenedTab();
		if (picker.syraFieldsCount) {
			var funcName;
			var templates = this.templatesTool.countTemplates[picker.syraFieldsCount];
			if (templates) {
				funcName = picker.syraFieldsCount;
			} else {
				templates = this.templatesTool.defaultTemplates;
				funcName = picker.syraTemplate;
			}
			var fields = designedItem.layoutContent.getFields();
			document.site.layoutUpdater.extractItems(designedItem.layoutContent, fields);
			this.templatesTool["build_" + funcName](designedItem, picker.syraTemplate, fields);
			return designedItem;
		}
		return null;
	},
	dispose: function() {
		if (this.itemTool) {
			this.itemTool.dispose();
		}
		if (this.domItem) {
			document.site.dom.removeChild(this.domItem);
		}
		if (this.editTitleField) {
			this.editTitleField.localizedSection = null;
		}
		this.editTitleField = this.itemTool = this.sectionModel = null;
		Designer.prototype.dispose.call(this);
	}
});
