// syracuse-ui/lib/tests/fields/binaryFieldTest.js
"use strict";
var helpers = require("syracuse-core/lib/helpers");
var FieldsTest = require("./fieldsTest").FieldsTest;

function BinaryFieldTest() {}

exports.BinaryFieldTest = helpers.defineClass(BinaryFieldTest, FieldsTest, {
	_initiliazeTestPage: function() {
		FieldsTest.prototype._initiliazeTestPage.call(this);
		syra_site.deltaManager.applyObjectDelta(this, this.$prototype.$properties, {
			"embededPDF": {
				$isTitleHidden: true,
				$type: "application/pdf",
				$url: "/syracuse-ui/lib/tests/binary/binarytest.pdf"
			},
			"embededPNG": {
				$isTitleHidden: true,
				$type: "image/png",
				$url: "/syracuse-ui/lib/tests/binary/binarytest.png"
			}
		}, true);
	},
	loadBox: function() {
		FieldsTest.prototype.loadBox.call(this, {});
		this.applyChange({
			embededPDF: {},
			embededPNG: {}
		});
	},
	_makeBlock: function($isEditMode) {
		return [this._makeEditReadBlock("PDF", [{
			$bind: "embededPDF"
		}]), this._makeEditReadBlock("PNG", [{
			$bind: "embededPNG"
		}])];
	}
});
