// syracuse-ui/lib/field/child/childFieldRecord.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Article = require("syracuse-ui/lib/article/article").Article;
var ContextMenu = require('syracuse-ui/lib/field/menus/contextMenu').ContextMenu;

function ChildFieldRecord(){
}

exports.ChildFieldRecord = helpers.defineClass(ChildFieldRecord, Article, {
    remove: function(){
        this.$$item.remove();
    },
    drawBox: function(){
        this.$$item = $(this.list._core);
        this._contextMenusSlot = document.createElement("div");
        this._contextMenusSlot.className = this.list.$skin + "-actions-cell";
        this.list._core.appendChild(this._contextMenusSlot);
        var card = this.list._core.appendChild(document.createElement("div"));
        card.className = this.list.$skin + "-record";
        document.itemFactory.load($(card), this.list.$item.$cards, this);
    },
    notifyChange: function($bind, value){
        if (!this.$isDisabled && !this.$isDeleted) {
            var sendBag = this.getPage().ensureSendBag(this.list);
            (sendBag.dataset = sendBag.dataset || {})[$bind] = value;
            this.list._notifyChangeToParent(sendBag.dataset);
        }
    },
    applyActionLinkChange: function(resources, record){
        Article.prototype.applyActionLinkChange.call(this, resources, record);
        if (resources && (resources.$links !== undefined || resources.$actions !== undefined)) {
            if (!this._contextMenu) {
                (this._contextMenu = new ChildFieldContextMenu()).create(this);
            }
            if (resources.$links === null && this._contextMenu) {
                this._contextMenu.clearMenuItems();
            }
        }
    },
    applyMetaData: function(metaData){
    },
    dispose: function(){
        delete this.list
        if (this._contextMenu) {
            document.controller.disposeObject(this._contextMenu);
            delete this._contextMenu;
        }
        Article.prototype.dispose.call(this);
    }
});


function ChildFieldContextMenu(){
}

ChildFieldContextMenu = helpers.defineClass(ChildFieldContextMenu, ContextMenu, {
    clearMenuItems: function(){
        this._isMenuBtnVisible = false;
        this.$skinMenus = "s-list-record-menus";
        this._menusBtnSlot.style.display = "none";
    },
    create: function(field){
        var self = this
        ContextMenu.prototype.create.call(this, field, {
            onBeforeAddMenuItem: function($menuItem){
                if (!self._isMenuBtnVisible) {
                    self._menusBtnSlot.style.display = "";
                    self._isMenuBtnVisible = true;
                    if ($menuItem.$bind == "$lazyload") {
                        return false;
                    }
                }
            }
        });
    },
    appendOpener: function(){
        this._menusBtn = document.createElement("a");
        this._menusBtn.className = "s-list-menus-btn s-close";
        var title = this.field.list._localize.mn_actions;
        $(this._menusBtn).text(title);
        this._menusBtn.setAttribute("title", title);
        
        this._isMenuBtnVisible = false;
        this._menusBtn.setAttribute("id", this.field.id + "-menu-btn");
        (this._menusBtnSlot = document.createElement("div")).className = "s-list-quick-actions-slot";
        this._menusBtnSlot.style.display = "none";
        return this.field._contextMenusSlot.appendChild(this._menusBtnSlot).appendChild(this._menusBtn);
    },
    removeOpener: function(){
        if (this._menusBtn) {
            document.site.emptyDom(this.field._contextMenusSlot);
            delete this._menusBtn;
        }
    }
});
