// syracuse-ui/lib/tests/array/coreStyleSettings.js
"use strict";
exports.$item = {
	$layout: {
		$layoutType: "row",
		$widths: "60,40",
		$items: [{
			$layoutType: "stack",
			"$items": [{
				"$format": "grid",
				"$isEditMode": true,
				"$bind": "xarray",
				"$selectMode": "single"

			}, {
				"$format": "grid",
				"$bind": "xarray",
				"$selectMode": "single"

			}, {
				"$format": "cards",
				"$isEditMode": true,
				"$bind": "xarray",
				"$selectMode": "single",
				"$cardsByRowCount": 3,
				"$layout": {
					$items: [{
						$category: "section",
						"$layout": {
							"$items": [{
								"$bind": "xstring"
							}, {
								"$bind": "xchoice"
							}, {
								"$bind": "xdate"
							}, {
								"$bind": "xboolean"
							}, {
								"$bind": "xdecimal"
							}]
						}
					}]
				}
			}, {
				"$format": "cards",
				"$bind": "xarray",
				"$selectMode": "single",
				"$cardsByRowCount": 3,
				"$layout": {
					$items: [{
						$category: "section",
						"$layout": {
							"$items": [{
								"$bind": "xstring"
							}, {
								"$bind": "xchoice"
							}, {
								"$bind": "xdate"
							}, {
								"$bind": "xboolean"
							}, {
								"$bind": "xdecimal"
							}]
						}
					}]
				}
			}]
		}, {
			"$layoutType": "tabs",
			"$items": [{
				"$category": "section",
				"$id": "editSection",
				"$title": "Edit",
				"$layout": {
					"$items": [{
						"$bind": "$field$DefaultDeltaMode",
						"$isEditMode": true,
						"$format": "$radios",
						"$choiceLayout": 1
					}]
				}
			}, {
				"$category": "section",
				"$title": "Json Delta",
				"$layout": {
					"$items": [{
						"$bind": "$fieldJsonDelta",
						"$isTitleHidden": true,
						"$isEditMode": true,
						"$rows": 30,
						$isAutoSizeDisabled: true,
						"$skin": "s-json-field"
					}]
				}
			}]

		}]
	}
};

exports.cases = [{
	$title: "$isReadOnly true",
	delta: {
		"xarray": [{
			"$uuid": "record0",
			"$isReadOnly": true
		}, {
			"$uuid": "record1",
			"$isReadOnly": true
		}, {
			"$uuid": "record2",
			"$isReadOnly": true
		}]
	}
}, {
	$title: "$isReadOnly false",
	delta: {
		"xarray": [{
			"$uuid": "record0",
			"$isReadOnly": false
		}, {
			"$uuid": "record1",
			"$isReadOnly": false
		}, {
			"$uuid": "record2",
			"$isReadOnly": false
		}]
	}
}, {
	$title: "$isDisabled true",
	delta: {
		"xarray": [{
			"$uuid": "record0",
			"$isDisabled": true
		}, {
			"$uuid": "record1",
			"$isDisabled": true
		}, {
			"$uuid": "record2",
			"$isDisabled": true
		}]
	}
}, {
	$title: "$isDisabled false",
	delta: {
		"xarray": [{
			"$uuid": "record0",
			"$isDisabled": false
		}, {
			"$uuid": "record1",
			"$isDisabled": false
		}, {
			"$uuid": "record2",
			"$isDisabled": false
		}]
	}
}, {
	$title: "info warning error",
	delta: {
		"xarray": [{
			"$uuid": "record0",
			"$fieldStyle": "cst_sty_info"
		}, {
			"$uuid": "record1",
			"$valueStyle": "cst_sty_warning"
		}, {
			"$uuid": "record2",
			"$valueStyle": "cst_sty_error"
		}]
	}
}, {
	$title: "highlight verylow low",
	delta: {
		"xarray": [{
			"$uuid": "record0",
			"$fieldStyle": "cst_sty_highlight"
		}, {
			"$uuid": "record1",
			"$valueStyle": "cst_sty_verylow"
		}, {
			"$uuid": "record2",
			"$valueStyle": "cst_sty_low"
		}]
	}
}, {
	$title: "medium high veryhigh",
	delta: {
		"xarray": [{
			"$uuid": "record0",
			"$fieldStyle": "cst_sty_medium"
		}, {
			"$uuid": "record1",
			"$valueStyle": "cst_sty_high"
		}, {
			"$uuid": "record2",
			"$valueStyle": "cst_sty_veryhigh"
		}]
	}
}, {
	$title: "strongdecreasing decreasing stable",
	delta: {
		"xarray": [{
			"$uuid": "record0",
			"$fieldStyle": "cst_sty_strongdecreasing"
		}, {
			"$uuid": "record1",
			"$valueStyle": "cst_sty_decreasing"
		}, {
			"$uuid": "record2",
			"$valueStyle": "cst_sty_stable"
		}]
	}
}, {
	$title: "increasing strongincreasing",
	delta: {
		"xarray": [{
			"$uuid": "record0",
			"$fieldStyle": "cst_sty_strongdecreasing"
		}, {
			"$uuid": "record1",
			"$valueStyle": "cst_sty_decreasing"
		}, {
			"$uuid": "record2"
		}]
	}
}, {
	$title: "default",
	delta: {
		"xarray": [{
			"$uuid": "record0",
			"$fieldStyle": "cst_sty_default"
		}, {
			"$uuid": "record1",
			"$valueStyle": "cst_sty_default"
		}, {
			"$uuid": "record2",
			"$valueStyle": "cst_sty_default"
		}]
	}
}];
