// syracuse-ui/lib/tests/array/graphTest.js
"use strict";
var helpers = require("syracuse-core/lib/helpers")
var MultiArrayTest = require("./multiArrayTest").MultiArrayTest;
var arraySettings = require("./arraySettings");
function GraphTest(){
}

exports.GraphTest = helpers.defineClass(GraphTest, MultiArrayTest, {
    _defineArticle: function(){
        this.$item = {
            $layout: {
                $layoutType: "columns",
                $layoutSubType: "60,40",
                $items: [{
                    $layoutType: "stack",
                    $items: [{
                        $bind: "xarray",
                        $chartType: "highCharts",
                        $style: "area"
                    }, {
                        $bind: "xarray",
                        $chartType: "highCharts",
                        $style: "areaspline"
                    }, {
                        $bind: "xarray",
                        $chartType: "highCharts",
                        $style: "bar"
                    }, {
                        $bind: "xarray",
                        $chartType: "highCharts",
                        $style: "column"
                    }, {
                        $bind: "xarray",
                        $chartType: "highCharts",
                        $style: "line"
                    }, {
                        $bind: "xarray",
                        $chartType: "highCharts",
                        $style: "pie"
                    }, {
                        $bind: "xarray",
                        $chartType: "highCharts",
                        $style: "scatter"
                    }, {
                        $bind: "xarray",
                        $chartType: "highCharts",
                        $style: "spline"
                    }]
                }, arraySettings.settingBar]
            }
        };
    },
    refreshList: function(records){
        if (records === undefined) {
            records = this.getResources().getCube();
        }
        this.applyChange({
            $properties: {
                "xarray": {
                    $totalResults: records.length
                }
            },
            $field$editMode: "array",
            "xarray": records
        });
    },
    _makeArrayPrototype: function(){
        var $prototype = MultiArrayTest.prototype._makeArrayPrototype.call(this);
        $prototype.$properties.xarray = {
            "$type": "application/x-array",
            "$axes": [{
                "$title": "Articles vendus",
                "$hierarchies": [["article_h"]]
            }, {
                "$title": "measures",
                "$hierarchies": [["$measures"]]
            }],
            "$cube": {
                "$mode": "table",
                "$style": "xy",
                "$displaysOneMeasure": false,
                "$title": "Coûts articles",
                "$hierarchies": {
                    "article_h": {
                        "$title": "Articles",
                        "$properties": ["ARTICLE"],
                        "$members": [],
                        "$isDefault": true
                    }
                },
                "$measures": {
                    "matiere_premiere": {
                        "$property": "matiere_premiere",
                        "$title": "Matière premiere",
                        "$mode": "sum",
                        "$style": "line",
                        "$isStacked": false,
                        "$isNormalized": false,
                        "$isDefault": true
                    },
                    "main_oeuvre": {
                        "$property": "main_oeuvre",
                        "$title": "Main d'oeuvre",
                        "$mode": "sum",
                        "$style": "stick",
                        "$isStacked": false,
                        "$isNormalized": false
                    },
                    "cout_machine": {
                        "$property": "cout_machine",
                        "$title": "Cout machine",
                        "$mode": "sum",
                        "$style": "spline",
                        "$isStacked": false,
                        "$isNormalized": false
                    }
                }
            },
            "$item": {
                "$properties": {
                    "ARTICLE": {
                        "$title": "Intitulé article",
                        "$type": "application/x-string"
                    },
                    "code": {
                        "$title": "Code article",
                        "$type": "application/x-string"
                    },
                    "matiere_premiere": {
                        "$title": "Matiere première",
                        "$type": "application/x-decimal"
                    },
                    "main_oeuvre": {
                        "$title": "Main d'oeuvre",
                        "$type": "application/x-decimal"
                    },
                    "cout_machine": {
                        "$title": "Coût machine",
                        "$type": "application/x-decimal"
                    }
                },
                "$links": {
                    "$details": {
                        "$title": "Details",
                        "$url": "{$baseUrl}/article('{$key}')",
                        "$type": "{$baseType}.article.$details"
                    },
                    "$edit": {
                        "$title": "Edit",
                        "$url": "{$url}/$workingCopies",
                        "$type": "{$baseType}.article.$edit",
                        "$method": "POST"
                    },
                    "$delete": {
                        "$title": "Delete",
                        "$url": "{$url}",
                        "$type": null,
                        "$method": "DELETE"
                    }
                }
            }
        };
        return $prototype;
        
    }
});
