// syracuse-ui/lib/tests/array/multiArrayTest.js
"use strict";
var helpers = require("syracuse-core/lib/helpers")
var FieldsTest = require("../fieldsTest").FieldsTest;
var ListAuthoring = require("syracuse-ui/lib/field/array/authoring/listAuthoring").ListAuthoring;
var convergenceStyle = require("./convergenceStyle");
var partialDeltaTests = require("./delta/partialDeltaTests");
var deltaTests = require("./delta/deltaTests");
var multiSettings = require("./multiSettings");
var multiTestCases = require("./multiTestCases");
var utility = require("./utility");
function MultiArrayTest(){
}

exports.MultiArrayTest = helpers.defineClass(MultiArrayTest, FieldsTest, {
    _defineArticle: function(){
        this.$item.$layout = {
            $layoutType: "columns",
            $layoutSubType: "60,40",
            $items: [multiTestCases.cases, multiSettings.settingBar]
        };
    },
    onFetchHandler: function(options){
        if (options && options.params) {
            delete options.field;
            this.notifyChangeToServer(options);
        }
    },
    notifyActionChange: function(target, value){
        if (value.$empty || value.$refresh) {
            if (value.$empty) {
                this.refreshList([]);
            }
            if (value.$refresh) {
                this.refreshList();
            }
            return;
        }
        FieldsTest.prototype.notifyActionChange.call(this, target, value);
    },
    notifyDataChange: function(field, value){
        var notify = true;
        switch (field.$item.$bind) {
            case "$field$PartialDelta":
                this.applyEdit(partialDeltaTests.cases[value]);
                break;
            case "$field$DefaultDeltaMode":
                this.applyEdit(deltaTests.cases[value]);
                break;
            case "$field$fixedBodyHeight":
                var numValue = parseInt(value || 0, 10);
                if (numValue) {
                    var min = Math.max(35, numValue) + "";
                    if (min != value) {
                        field.setDataValue(min);
                        value = min;
                    }
                }
                break;
            case "$fieldRecordCount":
                this.refreshList();
                notify = false;
                break;
            default:
                if (field.$item.$bind.indexOf("$field$capability") >= 0) {
                    this.applyChange({
                        $properties: {
                            xarray: utility.applyCapabilityChange(this, field.$item.$bind)
                        }
                    });
                    notify = false;
                }
                break;
        }
        if (notify) {
            FieldsTest.prototype.notifyDataChange.call(this, field, value);
        }
    },
    applyEdit: function($choice){
        //insert one Record
        var self = this;
        self.isDeltaMode = false; //allow erase of all list ( data + metadata on record)
        self.refreshList();
        self.isDeltaMode = true;
        var delta = $choice.delta;
        if (delta.$isPartialDelta) {
            self.$isFusionPage = true;
        }
        if (self.isDiagnosesPanelHidden()) {
            this.diagnosesPanel.disableViewer = true;
        }
        
        switch ($choice.$id) {
            case "reorder":
                if (delta.$isPartialDelta) {
                    var count = self.dataset.xarray.length - 1;
                    delta.xarray = self.dataset.xarray.map(function(record, index){
                        return {
                            $uuid: record.$uuid,
                            $index: count - index
                        };
                    }).reverse();
                }
                else {
                    delta.xarray = self.dataset.xarray.map(function(record){
                        return {
                            $uuid: record.$uuid
                        };
                    }).reverse();
                }
                break;
            case "deleteDefault":
                delta.xarray = self.dataset.xarray.filter(function(record){
                    return (record.$uuid != "record3") && (record.$uuid != "record6") && (record.$uuid != "record7");
                });
                break;
        }
        //self.boundFields.$fieldJsonDelta[0].setDataValue(JSON.stringify(delta, null, 2));
        self.applyChange({
            $fieldJsonDelta: JSON.stringify(delta, null, 2)
        });
        self.applyChange(delta, true);
        delete self.$isFusionPage;
        this.diagnosesPanel.disableViewer = false;
    },
    _initiliazeTestPage: function(){
        FieldsTest.prototype._initiliazeTestPage.call(this);
        this.deltaManager.applyObjectDelta(this.$prototype, this._makeArrayPrototype(), true);
    },
    loadBox: function(){
        var self = this;
        FieldsTest.prototype.loadBox.call(self, {
            $fieldRecordCount: 8,
            $field$fixedBodyHeight: 0,
            $field$fitContainer: false
        });
        //find first DeskttopKlist
        self.boundFields["xarray"].some(function(item){
            if (item.builder) {
                self.designedField = item;
            }
            return self.designedField != null;
        });
        /*self.idMap["$partialDeltaLinks"].addMenuItems(partialDeltaTests.cases.map(function($item, index){
         return {
         $value: index,
         $title: $item.$title
         };
         }), true);
         self.idMap["$defaultDeltaModeLinks"].addMenuItems(deltaTests.cases.map(function($item, index){
         return {
         $value: index,
         $title: $item.$title
         };
         }), true);*/
        self._appendAuthoring();
        self.refreshList();
        
        self.page.dataset.$etag = 1;
    },
    _appendAuthoring: function(){
        (this.authoring = new ListAuthoring()).loadBox(this);
        this.authoring.designedTestFields = this.boundFields["xarray"].filter(function(item){
            return item.builder != null;
        });
        this.idMap.authoringTest.$$body.empty().append(this.authoring.layoutContent.$$container)[0].style.backgroundColor = "#222222";
        this.authoring.layoutContent.$$container.attr("data-s-article", this.authoring.id);
    },
    _makeArrayPrototype: function(){
        return {
            "$properties": {
                $fieldJsonDelta: {
                    "$title": "delta",
                    "$type": "application/x-string"
                },
                "$field$PartialDelta": {
                    $title: "isPartialDelta",
                    "$type": "application/x-choice",
                    "$value": {
                        $type: "application/x-integer",
                        $enum: partialDeltaTests.cases.map(function($item, index){
                            return {
                                $value: index,
                                $title: $item.$title
                            };
                        })
                    }
                },
                "$field$DefaultDeltaMode": {
                    $title: "isDeltaMode",
                    "$type": "application/x-choice",
                    "$value": {
                        $type: "application/x-integer",
                        $enum: deltaTests.cases.map(function($item, index){
                            return {
                                $value: index,
                                $title: $item.$title
                            };
                        })
                    }
                },
                "$field$editMode": {
                    $title: "$editMode",
                    $type: "application/x-choice",
                    $isHidden: true,
                    $value: {
                        $type: "application/x-integer",
                        $enum: [{
                            $value: "array",
                            $title: "array (default)"
                        }, {
                            $value: "record",
                            $title: "record"
                        }, {
                            $value: "cell",
                            $title: "cell"
                        }]
                    }
                },
                
                $fieldRecordCount: {
                    $title: "Records",
                    "$type": "application/x-integer"
                },
                $field$fitContainer: {
                    $title: "$fitContainer",
                    "$type": "application/x-boolean"
                },
                $field$fixedBodyHeight: {
                    $title: "$fixedBodyHeight",
                    "$type": "application/x-integer"
                },
                "xarray": {
                    "$title": "xarray",
                    "$capabilities": "sort,filter,insert,append,delete,reorder",
                    "$type": "application/x-array",
                    "$itemsPerPage": 5,
                    "$item": {
                        "$descriptor": "prototype employee.$edit",
                        "$type": "application/json",
                        "$pluralType": "children",
                        "$representation": "employee",
                        "$properties": {
                            "xstring": {
                                "$title": "xstring",
                                "$type": "application/x-string",
                                
                                "$maxLength": 30,
                                "$capabilities": "sort,filter,alphaTab"
                            },
                            "xchoice": {
                                $title: "xchoice",
                                $type: "application/x-choice",
                                $value: {
                                    $type: "application/x-integer",
                                    $enum: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10].map(function(value){
                                        return {
                                            "$value": value,
                                            $title: "Value " + value
                                        };
                                    })
                                }
                            },
                            "xdate": {
                                $title: "xdate",
                                $type: "application/x-date",
                                "$capabilities": "sort,filter,alphaTab"
                            },
                            "xboolean": {
                                $title: "xboolean",
                                $type: "application/x-boolean",
                                "$capabilities": "sort,filter"
                            },
                            "xdecimal": {
                                $title: "xdecimal",
                                $type: "application/x-decimal",
                                "$capabilities": "sort,filter",
                                
                                "$isNullable": true
                            }
                        },
                        "$links": {},
                        "$actions": {}
                    },
                    "$actions": {}
                }
            },
            "$actions": {
                "$empty": {
                    "$title": "Empty list",
                    "$method": "PUT"
                },
                "$refresh": {
                    "$title": "Refresh list",
                    "$method": "PUT"
                }
            }
        };
    },
    getDefaultRecords: function(){
        var $fieldRecordCount = this.boundFields.$fieldRecordCount[0].getDataValue();
        var records = this.getResources().getRecords($fieldRecordCount);
        records.forEach(function(record){
            delete record.$index;
        });
        //this.boundFields.$field$dataset[0].setDataValue(JSON.stringify(records, null, 2));
        return records;
    },
    refreshList: function(records){
        if (records === undefined) {
            records = this.getDefaultRecords();
        }
        this.applyChange({
            $properties: {
                "xarray": {
                    $isReadOnly: false,
                    $isDisabled: false,
                    $totalResults: records.length
                }
            },
            $field$editMode: "array",
            "xarray": records
        });
    }
});
