// syracuse-ui/lib/tests/array/utility.js
"use strict";
var convergenceStyle = require("./convergenceStyle");

exports.addCapabilitySetting = function(testPage){
    testPage.deltaManager.applyObjectDelta(testPage.$prototype, {
        $properties: {
            "$field$capabilityInsert": {
                $type: "application/x-boolean",
                $title: "insert",
                $isMetaData: true
            },
            "$field$capabilityDelete": {
                $type: "application/x-boolean",
                $title: "delete",
                $isMetaData: true
            },
            "$field$capabilityAppend": {
                $type: "application/x-boolean",
                $title: "append",
                $isMetaData: true
            },
            "$field$capabilityReorder": {
                $type: "application/x-boolean",
                $title: "reorder",
                $isMetaData: true
            }
        },
        $field$capabilityInsert: true,
        $field$capabilityDelete: true,
        $field$capabilityAppend: true,
        $field$capabilityReorder: true
    }, true);
};

exports.applyCapabilityChange = function(testPage, $field$capability){
    var $capabilities = "";
    ["$field$capabilityInsert", "$field$capabilityDelete", "$field$capabilityAppend", "$field$capabilityReorder"].forEach(function($capability){
        if (testPage.boundFields[$capability][0].getDataValue()) {
            $capabilities += $capability.replace("$field$capability", "").toLowerCase() + ",";
        }
    });
    if ($capabilities.length > 0) {
        $capabilities = $capabilities.slice(0, $capabilities.length - 1);
    }
    return {
        $capabilities: $capabilities
    };
};

function _buildStyleChoiceList(){
    return Object.keys(convergenceStyle.$styles).map(function($key){
        return {
            $value: $key,
            $title: $key
        }
    });
};
/*
 exports.addStyleSetting = function(testPage){
 var choiceList = _buildStyleChoiceList();
 testPage.deltaManager.applyObjectDelta(testPage.$prototype, {
 $properties: {
 "$field$fieldStyle": {
 $title: "$fieldStyle",
 $type: "application/x-choice",
 $value: {
 $type: "application/x-integer",
 $enum: choiceList
 }
 },
 "$field$fieldStyleDesc": {
 $type: "application/x-string"
 },
 "$field$titleStyle": {
 $title: "$titleStyle",
 $type: "application/x-choice",
 $value: {
 $type: "application/x-integer",
 $enum: choiceList
 }
 },
 "$field$titleStyleDesc": {
 $type: "application/x-string"
 },
 "$field$valueStyle": {
 $title: "$valueStyle",
 $type: "application/x-choice",
 $value: {
 $type: "application/x-integer",
 $enum: choiceList
 }
 },
 "$field$valueStyleDesc": {
 $type: "application/x-string"
 }
 },
 $field$fieldStyle: choiceList[0].$value,
 $field$titleStyle: choiceList[0].$value,
 $field$valueStyle: choiceList[0].$value
 }, true);
 };
 exports.onStyleChange = function(testPage, $bind, value){
 var desc = {};
 desc[$bind + "Desc"] = convergenceStyle.$styles[value].$css;
 testPage.applyChange(desc);
 };*/
