// syracuse-ui/lib/fusion/core/client/eventhandling/formula.js
"use strict";

/*
 Copyright (c)2012 Sage.
 http://wwww.sage.com
 */
/**
 * @fileoverview Fusion "Formula wizard" windows event handling class
 */
var util = require('syracuse-ui/lib/fusion/tools/util');
var evtTools = require('../eventhandling/common').CommonEventHandlingTools;
var requestActions = require('../sap/srvactions');
var keys = require('syracuse-ui/lib/fusion/tools/constant').keybordKey;


var _evtProcFormula = {};
exports.FormulaEventHandling = _evtProcFormula;

/* "Formula wizard dialogs" events handling class */
_evtProcFormula.sap = function() {
	this.eventListeners = function() {
		return [["wdgt.grid.selline", selectLine, true, true], ["wdgt.field.keydown", keyDown, true, true], ["wdgt.field.keyup", keyUp, true, true], ["wdgt.field.mousedown", onFieldMD, true, true], ["wdgt.field.click", click, true, true], ["wdgt.link.click", onLinkClick, true, true], ["wdgt.record.picker", clickRecordPicker, true, true], ["wdgt.win.close", onClose, true, true], ["wdgt.win.closeall", onCloseAll, true, true], ["wdgt.win.validate", onValidate, true, true], ["wdgt.field.escape", applyEscape, true, true], ["wdgt.field.applyShortCut", applyShortCut, true, true]];
	};

	function _getActParams(winFormula) {
		var fld;
		return {
			"v": winFormula && (fld = winFormula._formulaFld) ? fld.getValue() : "",
			"xType": syra_fusion.syraUtil.dataTypes.stringType,
			"xFmt": syra_fusion.sapUtil.fieldModeType.edit
		};
	}

	function clickRecordPicker(event) {
		return undefined;
	}

	function onFieldMD(event) {
		return undefined;
	}

	function click(event) {
		var position = event.target.getCaretPosition && event.target.getCaretPosition();
		if (position >= 0) {
			event.target.caretPosition = position;
			event.target.selectionEnd = event.target.input.selectionEnd;
			event.target.selectionStart = event.target.input.selectionStart;
		}
		return true;
	}

	function helpUrl(event, fieldInfo) {
		var found = _findItem(fieldInfo, "help");
		if (found) {
			if (found.$field.$location.helpKey) {
				event.result = {
					"$act": requestActions.svcHelp,
					"post": {
						"callback": help,
						"args": [found]
					},
				};
				event.opts = {
					"$link": "$4glHelp",
					"$keyword": found.$field.$location.helpKey,
					"$title": ""
				};
			}
		}
		return true;
	}

	function help(url, target, show) {
		var metaData = {};
		//if (show !== undefined) metaData.stt = (show) ? 458711 : 456611;
		target.setValue(url, metaData, null);
	}

	function onLinkClick(event) {
		//TODO : à factorise avec grid, list, ...
		var act, $menu = event.target.$item;
		if ((act = event.target.$act) || (act = $menu && $menu.$act)) {
			// Action should be an expression...
			if (typeof act == 'string' && act.indexOf("{") >= 0 && (actEx = _getActionId(act))) {
				act = actEx;
			}
			// Native link
			event.result = {
				"$act": act,
				"params": _getActParams(this.winModel)
			};
			event.opts = {
				"$link": (event.target.$sourceBind || event.target.$bind)
			};
		}
		return true;
	};

	function onClose(event) {
		event.result = {
			$act: requestActions.getSrvAction("SRV_ABANDON"),
			params: {
				v: ""
			}
		};
		return true;
	};

	function onCloseAll(event) {
		var wins = [];
		for (var ii = (event.target.sheets.length - 1); ii > 0; ii--) {
			wins.push( /*event.target.sheets[ii].$fusionPageMeta.winId*/ null);
		}
		if (wins.length == 0)
			wins.push(null);
		event.result = {
			"$act": requestActions.getSrvActionEX("SUP_WINS_CLOSE"),
			"params": {
				"sudo": wins,
				"v": ""
			},
			post: {
				"callback": function() {
					event.callback();
				},
				"onErrCallBack": function() {
					event.callback();
				},
				"args": [this, event]
			},

		};
		return true;
	}

	function onValidate(event) {
		event.result = {
			"$act": requestActions.getSrvAction("SRV_CHAMP_SUIVANT"),
			"params": _getActParams(this.winModel)
		};
		return true;
	}

	function applyEscape(opt) {
		return true;
	}

	function applyShortCut(opt) {
		if (opt.data.shortcuts.enter) {
			return this.winModel && event && event.target && event.target == this.winModel._formulaFld ? false : true;
		} else
		if (opt.data.shortcuts.tab) {
			return true;
		}
		return false;
	}

	function keyDown(event) {
		return false;
	};

	function keyUp(event) {
		var position = event.target.getCaretPosition && event.target.getCaretPosition();
		if (position >= 0) {
			event.target.caretPosition = position;
			event.target.selectionEnd = event.target.input.selectionEnd;
			event.target.selectionStart = event.target.input.selectionStart;
		}
		return false;
	};

	function selectLine(event) {
		var fieldInfo = _getFieldInfo(event),
			found;
		if (fieldInfo.isSpecialNode) {
			event.result = {
				$act: null,
				params: {
					target: {}
				}
			};
			// send a 1071 action
			event.result.$act = requestActions.getSrvAction("SRV_ELEMENT_FORMULE");
			event.result.params.v = fieldInfo.$key;
			return true;
		} else {
			var itemForm = this.winModel._formulaFld;
			itemForm.formulaFull = itemForm.getValue();
			if (itemForm.boxParent.$opened) {
				if (!itemForm.formulaFull) {
					itemForm.formulaFull = fieldInfo.formula;
					itemForm.caretPosition = itemForm.formulaFull;
				} else {
					if (itemForm.formulaFull.indexOf('%') !== -1)
						itemForm.formulaFull = itemForm.formulaFull.replace('%', fieldInfo.formula);
					else
						itemForm.formulaFull = itemForm.formulaFull.substr(0, Math.min(itemForm.caretPosition, itemForm.selectionStart)) + fieldInfo.formula + itemForm.formulaFull.substr(Math.max(itemForm.caretPosition, (itemForm.selectionEnd ? itemForm.selectionEnd : 0)));
				}
				itemForm.setInputValue(itemForm.formulaFull.indexOf('%') ? itemForm.formulaFull.replace('%', '') : itemForm.formulaFull);
				itemForm.caretPosition = itemForm.formulaFull.length;
				itemForm.selectionEnd = itemForm.caretPosition;
				itemForm.selectionStart = itemForm.caretPosition;
			}
			if (fieldInfo.helpKey && (found = _findItem(fieldInfo, "help"))) {
				found.$field.$location.helpKey = fieldInfo.helpKey;
				helpUrl(event, fieldInfo);
			}
			return true;
		}
	};

	function _findItem(fieldInfo, name) {
		var keys = Object.keys(fieldInfo.winModel.syraModel.idMap);
		for (var i = 0; i < keys.length; i++) {
			var item = fieldInfo.winModel.syraModel.idMap[keys[i]];
			if (item.$item && item.$item.$bind === name) {
				return item;
			}
		}
		return null;
	}

	function _getFieldInfo(event) {
		var nl = (event.data && event.data.line) ? event.data.line : 0,
			fieldInfo = evtTools.getFieldInfo(syra_fusion.syraUtil.getXElement(event.target), nl),
			$fusionPageMeta = syra_fusion.syraUtil.getFusionPageMeta(event.target),
			record, treeNode = null;
		fieldInfo.isLeaf = true;
		fieldInfo.nl = nl;
		fieldInfo.xid = syra_fusion.syraUtil.getMetaFromObject(syra_fusion.syraUtil.getXElement(event.target)).$bind;
		//fieldInfo.blockIst = syra_fusion.sapUtil.makeIst(fieldInfo.fieldIst.win, syra_fusion.sapUtil.getBlockInst(fieldInfo.xid), 0);
		fieldInfo.data = event.data;
		fieldInfo.array = syra_fusion.syraUtil.getXElementGrid(event.target);
		fieldInfo.target = event.target;
		var recTarget = (fieldInfo.data.uuid !== undefined ? fieldInfo.array.findDataRecord(fieldInfo.data.uuid) : undefined);
		if (recTarget) {
			var srvIdx = recTarget.dataRecordIndex;
			record = syra_item.findRecordByServerIndex(fieldInfo.array, srvIdx);
			if (record && (treeNode = record.treeNode) && treeNode.children) {
				fieldInfo.isLeaf = (treeNode.children.length == 0);
			}
			if ((!fieldInfo.isLeaf || treeNode.parentId == "") && recTarget.dataRecord[fieldInfo.xid + "2"]) {
				fieldInfo.isSpecialNode = true;
				fieldInfo.$key = recTarget.dataRecord[fieldInfo.xid + "2"];
			} else {
				fieldInfo.formula = recTarget.dataRecord[fieldInfo.xid + "6"];
				fieldInfo.helpKey = (fieldInfo.xid == "fl" ? "" : recTarget.dataRecord[fieldInfo.xid + "5"]);
			}
		}
		return fieldInfo;
	};
	//TODO : à factorise avec Link  ...

	function _getActionId(actionExpr) {
		var actStrg, actTab, ret = null,
			matches = actionExpr.match(/(\{.*?\})/g);
		if (matches && matches.length > 0) {
			if (matches[0].indexOf("@") >= 0) {
				actStrg = matches[0].substr(1, matches[0].length - 2);
				if ((actTab = actStrg.split("@")) && actTab.length > 0) {
					ret = requestActions.getActionByCategorie(actTab[0], actTab[1]);
				}
			} else {
				// Others expression patterns...
			}
		}
		return ret;
	}
};
