// syracuse-ui/lib/fusion/core/client/eventhandling/grid.js
"use strict";

/*
 Copyright (c)2012 Sage.
 http://wwww.sage.com
 */
/**
 * @fileoverview Fusion Grid event handling classes
 */
/*
 * ["wdgt.grid.clickline", focusLine, true, true],
 ["wdgt.cell.click", clickCell, true, true],
 ["wdgt.grid.clickcellCheck", focusCellCheckBox, true, true],
 ["wdgt.grid.clickcellIco", focusCellIco, true, true],
 */
var sapUtil = require('syracuse-ui/lib/fusion/core/client/sap/util').Fusion;
var syraUtil = require('syracuse-ui/lib/fusion/core/client/sap/util').Syra;
var requestActions = require('../sap/srvactions');
var util = require('syracuse-ui/lib/fusion/tools/util');
var svcFmt = require('syracuse-ui/lib/fusion/core/tools/format').svcFormatter;

var _evtProcGrid = {};
exports.GridEventHandling = _evtProcGrid;

/* "Server page input" event handling class for "Grid" widget */
_evtProcGrid.sap = function(){
    this.eventListeners = function(){
        return [["wdgt.cell.click", clickCell, true, true], ["wdgt.grid.selline", selectLine, true, true], ["wdgt.record.action", actionLine, true, true], ["wdgt.array.action", actionArray, true, true], ["wdgt.cell.picker", clickPicker, true, true], ["wdgt.cell.calendarclosed", onDatePickerClose, true, true]];
    };
    function actionLine(event){
        var processed = undefined, nl, xid, currIst, isEdit, newIst, firstFocus, changeLine, acts = {}, $fusionPageMeta, srvActEx, srvAct;
        if (event.target && ($fusionPageMeta = syraUtil.getFusionPageMeta(event.target))) {
            nl = parseInt(event.target.getArticle().$recordIndex, 10) + 1;
            xid = syraUtil.getMetaFromObject(event.target.getArticle().getArticleParent()).$bind;
            currIst = this.sapController.getCurrInst();
            newIst = sapUtil.makeIst($fusionPageMeta.winModel.getWinId(), xid, nl);
            firstFocus = !sapUtil.cmpIstBlock(currIst, newIst);
            changeLine = (!firstFocus && currIst.nl != nl)
            isEdit = (!firstFocus && !changeLine && currIst.edit)
            
            if (firstFocus) {
                /*************************** 
                 * hack temporaire : objectif trouver la premiere colonne sur laquelle l'utilisateur peut cliquer
                 */
                xid = event.target.getArticle().getArticleParent().$item.$layout.$items[0].$bind
                /*************************************** 
                 *
                 */
                newIst = sapUtil.makeIst($fusionPageMeta.winModel.getWinId(), xid, nl);
            }
            else {
                newIst = sapUtil.makeIst($fusionPageMeta.winModel.getWinId(), currIst.xid, nl);
            }
            switch (event.target.$bind) {
                case "$delete":
                    srvActEx = "SUP_LINE_DELETE";
                    srvAct = "SRV_DELETELINEGRID";
                    break;
                case "$create":
                    srvActEx = "SUP_LINE_INSERT";
                    srvAct = "SRV_INSERTLINEGRID";
                    break;
            }
            if (firstFocus || changeLine || isEdit) { // focus not already on this table or line change same ligne but edit field
                acts.srvAct = requestActions.getSrvActionEX(srvActEx);
                acts.send = true;
            }
            else {
                acts.srvAct = requestActions.getSrvAction(srvAct);
                acts.send = true;
            }
            if (acts.send) {
                event.result = {
                    "$act": acts.srvAct
                };
                event.result.params = {
                    "target": newIst
                }
                event.result.post = acts.post
                processed = true;
            }
        }
        return true;
    };
    function actionArray(event){
        var processed = undefined, nl, xid, currIst, newIst, firstFocus, changeLine, acts = {}, $fusionPageMeta, srvActEx, srvAct;
        var sIndex = event.target.getArticle().getSelectedRecordsIndex();
        if (event.target && ($fusionPageMeta = syraUtil.getFusionPageMeta(event.target))) {
        }
        return true;
    };
    function selectLine(event){
        var processed = undefined, fieldInfo = {}, $fusionPageMeta, newIst, acts = {}, targetType;
        if (event.target && ($fusionPageMeta = syraUtil.getFusionPageMeta(event.target))) {
            fieldInfo.nl = event.data.line;
            fieldInfo.xid = syraUtil.getMetaFromObject(event.target).$bind;
            fieldInfo.currIst = this.sapController.getCurrInst();
            newIst = sapUtil.makeIst($fusionPageMeta.winModel.getWinId(), fieldInfo.xid, fieldInfo.nl);
            fieldInfo.firstFocus = !sapUtil.cmpIstBlock(fieldInfo.currIst, newIst);
            if (fieldInfo.firstFocus) { // focus not already on this table 
                acts.srvAct = requestActions.getSrvAction("SRV_COMMANDGRID"); // ask for focus 
                acts.send = true;
            }
            if (acts.send) {
                event.result = {
                    "$act": acts.srvAct
                };
                event.result.params = {
                    "target": newIst
                }
                event.result.post = acts.post
            }
        }
        return false;
    };
    function clickPicker(event){
        var process = undefined, pickerType;
        if (event.data) {
            if ((pickerType = event.data.pickerType)) {
                switch (pickerType) {
                    case "menus":
                        break;
                    case "tunnel":
                        break;
                    case "lookup":
                        break;
                    case "date":
                        process = clickCell.call(this, event);
                        process = datePickerClick.call(this, event, process);
                        break;
                    default:
                        process = clickCell.call(this, event)
                        break;
                }
            }
        }
        return process
    };
    function datePickerClick(event, process){
        if (process && !syraUtil.getPopupDateFromField(event.target)) {
            event.result.post = {
                "callback": function(evtHandlingObj, initialEvent, callerField){
                    var fieldContext = evtHandlingObj.sapController.getCurrInst();
                    _openDatePicker(true, fieldContext.raw ? svcFmt.formReply(fieldContext.raw, fieldContext.fmt) : null, evtHandlingObj, initialEvent, callerField);
                },
                "args": [this, event, event.target]
            };
        }
        else 
            if (!syraUtil.getPopupDateFromField(event.target)) {
                var fieldContext = this.sapController.getCurrInst();
                var currValue = syraUtil.getFusionPageMeta(event.target).winModel.getValue(syraUtil.getMetaFromObject(event.target).$bind, event.target.getArticle().$recordIndex + 1);
                if (currValue.v) {
                    event.result = {
                        "$act": requestActions.svcFormat
                    };
                    event.result.params = svcFmt.buildRqst(svcFmt.fmtSvc.toRaw, currValue.v, fieldContext.fmt, syraUtil.getFieldType(event.target), fieldContext);
                    event.result.post = {
                        "callback": _openDatePicker,
                        "args": [this, event, event.target]
                    };
                    process = true;
                }
                else {
                    var self = this
                    setTimeout(function(){
                        _openDatePicker(true, null, self, event, event.target);
                    }, 10);
                    process = true;
                }
            }
        return process;
    };
    function onDatePickerClose(event){
        var processed = undefined, fieldInfo;
        if (event.target && (fieldInfo = _getFieldInfo.call(this, event)) && fieldInfo.focus) {
            event.result = {
                "$act": requestActions.svcFormat
            };
            event.result.params = svcFmt.buildRqst(svcFmt.fmtSvc.rawToEdit, event.data.internalValue, fieldInfo.currIst.fmt, fieldInfo.targetType, fieldInfo.currIst);
            event.result.post = {
                "callback": function(succeeded, reply, evtHandlingObj, initialEvent){
                    if (fieldInfo.focus) {
                        syraUtil.getFusionPageMeta(initialEvent.target).winModel.setCurrCtxPopup(null);
                        initialEvent.data.doStatements({
                            "value": svcFmt.getReplyResult(reply)
                        });
                        delete initialEvent.doStatements;
                    }
                },
                "args": [this, event, event.target, fieldInfo]
            };
            processed = true;
        }
        return processed;
    };
    function _openDatePicker(succeeded, result, evtHandlingObj, initialEvent, field){
        var session, $fusionPageMeta = syraUtil.getFusionPageMeta(field);
        if (!syraUtil.getPopupDateFromField(field)) {
            if (succeeded) {
                initialEvent.data.doEvent(result ? {
                    "internalValue": parseInt(svcFmt.getReplyResult(result), 10)
                } : null);
                $fusionPageMeta.winModel.setCurrCtxPopup(field);
            }
            else {
                session = (evtHandlingObj.sapController.getSiteController()).getSession();
                $fusionPageMeta.winModel._showFieldError(field, [session.getlabel("error", "22", "Date entered is invalid; calendar can not be opened")]);
            }
        }
        delete initialEvent.data.doEvent;
    };
    function clickCell(event){
        // "target":... "type":... "data": {"nativeEvt:..."}
        var processed = undefined, fieldInfo = _getFieldInfo.call(this, event), acts;
        if (event.target) {
            switch (fieldInfo.targetType) {
                case syraUtil.dataTypes.booleanType:
                    acts = clickCellBoolean.call(this, fieldInfo, fieldInfo.newIst);
                    break;
                case syraUtil.dataTypes.iconType:
                    acts = clickCellIcon.call(this, fieldInfo);
                    break;
                case syraUtil.dataTypes.dateType:
                    acts = clickCellDefault.call(this, fieldInfo);
                    break;
                default:
                    acts = clickCellDefault.call(this, fieldInfo);
                    if (fieldInfo.data.doEvent) {
                        if (acts.send) {
                            acts.post = {
                                "callback": fieldInfo.data.doEvent,
                                "args": []
                            };
                        }
                    }
                    break;
            }
            if (acts.send) {
                event.result = {
                    "$act": acts.srvAct
                };
                event.result.params = {
                    "target": fieldInfo.newIst
                }
                event.result.post = acts.post
                processed = true;
            }
        }
        return processed;
    }
    function clickCellDefault(fieldInfo){
        var acts = {
            "send": false,
            "srvAct": null,
            "post": null
        };
        if (!fieldInfo.focus) { // focus change
            if (fieldInfo.target.$isReadOnly) {
                acts.srvAct = requestActions.getSrvAction("SRV_COMMANDGRID"); // ask for focus
                acts.send = true;
            }
            else 
                if (!fieldInfo.firstFocus) { // focus already on this table
                    acts.srvAct = requestActions.getSrvAction("SRV_GETFOCUSCELL"); // ask for editing
                    acts.send = true;
                }
                else {
                    // focus not on this table
                    acts.srvAct = requestActions.getSrvActionEX("SUP_CELL_EDIT"); //774 SRV_COMMANDGRID + SRV_GETFOCUSCELL
                    acts.send = true;
                }
        }
        return acts;
    };
    function clickCellBoolean(fieldInfo, newIst){
        var acts = clickCellDefault.call(this, fieldInfo);
        if (!acts.srvAct) {
            if (!fieldInfo.target.$isReadOnly) { // focus on this field
                acts.srvAct = requestActions.getSrvAction("SRV_GETFOCUSCELL"); // SRV_GETFOCUSCELL + SRV_GETFOCUSCELL
                acts.send = true;
            }
        }
        var newVal = (fieldInfo.target.currentValue == sapUtil.check.off ? sapUtil.check.on : sapUtil.check.off);
        var actionNext = (acts.srvAct == requestActions.getSrvActionEX("SUP_CELL_EDIT") || (!fieldInfo.firstFocus && !fieldInfo.currIst.edit)) ? requestActions.getSrvAction("SRV_COMMANDGRID") : requestActions.getSrvAction("SRV_GETFOCUSCELL")
        acts.post = {
            "exposeReply": true,
            "callback": toggleFieldAfterClick,
            "args": [this, fieldInfo.target, newIst, newVal, actionNext]
        };
        return acts;
    };
    function clickCellIcon(fieldInfo){
        var acts = {
            "send": false,
            "srvAct": null,
            "post": null
        };
        if (!fieldInfo.focus) { // focus change
            if (!fieldInfo.firstFocus) { // focus already on this table
                acts.srvAct = requestActions.getSrvActionEX("SUP_CELL_ICOCLICKFOCUS"); // 776 SRV_GETFOCUSCELL + SRV_KIB
                acts.send = true;
            }
            else {
                // focus not on this table
                acts.srvAct = requestActions.getSrvActionEX("SUP_CELL_ICOCLICK"); //775 SRV_COMMANDGRID + SRV_GETFOCUSCELL + SRV_KIB
                acts.send = true;
            }
        }
        else {
            acts.srvAct = requestActions.getSrvAction("SRV_KIB"); //
            acts.send = true;
        }
        return acts;
    }
    
    function toggleFieldAfterClick(reply, evtHandlingObj, $field, fieldIst, newValue, srvActChaine){
        var fieldValue = evtHandlingObj.sapController.getEntityValueInReply(reply, fieldIst), value, chain = null;
        if (fieldValue == null && fieldValue == undefined) {
            // Ok, field value has not been updated by the server after focus/click, we can process post local update...
            value = util.duplicateObj(fieldIst);
            value.v = newValue;
            evtHandlingObj.sapController.forceValue(value, false, true);
            // ...and finally we inform the server by this change
            if (srvActChaine) {
                chain = {
                    "$act": srvActChaine,
                    "params": {
                        "target": fieldIst
                    },
                    "post": null
                }
            }
        }
        return chain;
    }
    function _getFieldInfo(event){
        var fieldInfo = {}, $fusionPageMeta = syraUtil.getFusionPageMeta(event.target);
        fieldInfo.data = event.data;
        fieldInfo.nl = event.target.getArticle().$recordIndex + 1;
        fieldInfo.xid = syraUtil.getMetaFromObject(event.target).$bind;
        fieldInfo.currIst = this.sapController.getCurrInst();
        fieldInfo.newIst = sapUtil.makeIst($fusionPageMeta.winModel.getWinId(), fieldInfo.xid, fieldInfo.nl);
        fieldInfo.focus = sapUtil.cmpIst(fieldInfo.currIst, fieldInfo.newIst);
        fieldInfo.firstFocus = !sapUtil.cmpIstBlock(fieldInfo.currIst, fieldInfo.newIst);
        fieldInfo.target = event.target;
        fieldInfo.targetType = syraUtil.getFieldType(event.target);
        fieldInfo.readOnlyByModel = syraUtil.isFieldModelReadOnly($fusionPageMeta.winModel.getProto(), null, fieldInfo.xid);
        fieldInfo.changeLine = (!fieldInfo.firstFocus && fieldInfo.currIst.nl != fieldInfo.nl);
        fieldInfo.isEdit = (!fieldInfo.firstFocus && !fieldInfo.changeLine && fieldInfo.currIst.edit);
        return fieldInfo;
    }
};
