// syracuse-ui/lib/fusion/core/client/eventhandling/list.js
"use strict";

/*
 Copyright (c)2012 Sage.
 http://wwww.sage.com
 */
/**
 * @fileoverview Fusion "Left List" event handling classes
 */


var sapUtil = require('syracuse-ui/lib/fusion/core/client/sap/util').Fusion;
var syraUtil = require('syracuse-ui/lib/fusion/core/client/sap/util').Syra;
var requestActions = require('../sap/srvactions');

var _evtProcList = {};
exports.ListEventHandling = _evtProcList;

function getchangeLineActMap (winModel) {
	// Right now, only lists embedded in Selbox and Choose windows have a particular behaviour...
	var winName = winModel ? winModel.getWinName() : null, actSelLine = "SRV_BRWCHANGELINE", actSelLineSup = "SUP_LIST_SELLINE", forceSel = false;
	if(winName && winName === sapUtil.specialWindow.selbox.id || winName === sapUtil.specialWindow.choose.id) {
		forceSel = true;
		actSelLine = "SRV_BRWSELECTLINE";
		actSelLineSup = "SUP_LIST_SELLINEEX";
	}
	return [forceSel, requestActions.getSrvAction(actSelLine), requestActions.getSrvActionEX(actSelLineSup)];
}

/* "Server page input" event handling class for "Left List" widget */
_evtProcList.sap = function () {
	this.eventListeners = function () {
		return [["wdgt.list.toggle", beforeToggle, true, true],
				["wdgt.list.selline", changeLine, true, true],
				["wdgt.list.toggled", afterToggle]
		];
	};
	
	function beforeToggle(event, listXid) {
		var processed = undefined, currIst, $fusionPageMeta, listIst;
        if (event.target && event.data.open && ($fusionPageMeta = syraUtil.getFusionPageMeta(event.target))) {
 			currIst = this.sapController.getCurrInst();
 			listIst = sapUtil.makeIst($fusionPageMeta.winModel.getWinId(), listXid, currIst.nl);
 			if(!sapUtil.cmpIst(currIst, listIst)) {
				event.data.doEvent()
	            event.result = {"$act": requestActions.getSrvAction("SRV_GETFOCUS")};
	            event.result.params = {"target": listIst};
				/*
				 Bp: Lorsqu'on click sur le tiroir d'une liste gauche pour la premiere fois, les données sont bindées avant que 
				 L'objet grind ne soit créé.
				 J'execute donc le doEvent avant d'envoyer l'action getfocus.
				 cela peut peut-être posé un bp (à voir) si le serveur refuse le posItionnement sur cette liste (Je ne suis pas certain que le cas existe ?)
				 Il y a sans peut-être mieux à faire.
				*/
	            //event.result.post = {"callback": event.data.doEvent, "args": []};
	            processed = true; 				
 			}
        }
		return processed;
	}
	
	function afterToggle(event, listXid) {
        return undefined;
	}

	function changeLine(event) {
		var processed = undefined, currIst, $fusionPageMeta, listIst, srvAct, focus, actionsMap;
        if (event.target && ($fusionPageMeta = syraUtil.getFusionPageMeta(event.target))) {
 			currIst = this.sapController.getCurrInst();
 			listIst = sapUtil.makeIst($fusionPageMeta.winModel.getWinId(), event.data.xid, currIst.nl);
 			focus = sapUtil.cmpIst(currIst, listIst);
			actionsMap = getchangeLineActMap($fusionPageMeta.winModel);
 			if(!focus || currIst.nl != event.data.line || actionsMap[0]) {
 				listIst.nl = event.data.line;
	 			srvAct = focus ? actionsMap[1] : actionsMap[2];
	            event.result = {"$act": srvAct};
	            event.result.params = {"target": listIst};
	            processed = true;
	        }				
        }
        return processed;
	}
};
