// syracuse-ui/lib/field/array/filter/filterCell.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function FilterCell() {}

exports.FilterCell = helpers.defineClass(FilterCell, null, {
	addOperator: function() {
		var self = this;
		if (self.field.input) {
			self.field.input.className = self.field.input.className.replace("s-inplace-input", "s-filter-input");
		}
		self._operator = document.createElement("a");
		self._operator.className = "s-list-filter-cell-picker";
		self._operator.style.backgroundImage = "url('" + document.site.$item.$iconPath + "list/s-filter-none.png')";
		$(self.field._dataValue.appendChild(self._operator)).bind("click", function(event) {
			var options = {
				operator: self.operator,
				list: self.record.list,
				event: event,
				type: "operatorclick",
				doEvent: function() {
					if (!self.listPopup) {
						document.site.dialogManager.closePopups();
						self.$$list.empty();
						self._ensureFilterField();
						var operators = document.site.filterMaker.getFieldOperators(null, self.field, self.$filterField);
						for (var ii = 0, jj = operators.length; ii < jj; ii++) {
							var op = operators[ii];
							var link = document.createElement("a");
							link.className = "s-list-filter-item";
							link.setAttribute("data-s-op", op);
							if (self.operator == op) {
								link.className += " s-selected";
							}
							link.style.backgroundImage = "url('" + document.site.$item.$iconPath + "list/s-filter-" + op + ".png')";
							self.$$list[0].appendChild(link).textContent = self.record.boxParent.localize["flFilter_" + op];
						}
						self.listPopup = document.site.dialogManager.openPopup(self.record.boxParent, {
							content: self.record,
							slot: self.$$list[0],
							position: {
								my: "left top",
								at: "left bottom",
								of: $(self.layoutSlot)
							},
							onClose: function() {
								self.listPopup = null;
							}
						});
					} else {
						self.listPopup.close();
					}
				}
			};
			self.record.page.externalAdapter.onFilterEvent(options);
			return false;
		});
		if (!self.$$list) {
			var list = document.createElement("div");
			list.className = "s-list-filter-list";
			list.style.display = "none";
			self.$$list = $(self.field._dataValue.appendChild(list)).delegate("a.s-list-filter-item", "click", function(event) {
				self.setOperator($(this).attr("data-s-op"));
				self.listPopup.close();
				self.record.page.externalAdapter.onFilterEvent({
					operator: self.operator,
					list: self.record.list,
					filter: self,
					event: event,
					type: "operatorchange",
					doEvent: function() {
						var highField = self.record.boundFields[self.$bind + "$High"];
						if (self.operator === "between") {
							if (highField) {
								$(highField[0]._dataValue).toggle(true);
							} else {
								self.loadField(self.$field, self.$bind + "$High");
							}
						} else {
							if (highField) {
								$(highField[0]._dataValue).toggle(false);
							}
							var value = self.record.dataset[self.$bind];
							if (self.operator != "none") {
								if (value != null && value != "") {
									if (self.operator == "empty" || self.operator == "notempty") {
										var field = self.record.boundFields[self.$bind][0];
										field.setDataValue(null);
										self.record._buildFilter();
									} else {
										self.record._buildFilter();
									}
								} else {
									if (self.operator == "empty" || self.operator == "notempty") {
										self.record._buildFilter();
									}
								}
							} else {
								if (value != null && value != "") {
									var field = self.record.boundFields[self.$bind][0];
									field.setDataValue(null);
									field.notifyFieldChange(null);
								} else {
									self.record._buildFilter();
								}
							}
						}
					}
				});
			});
		}
	},
	_ensureFilterField: function() {
		if (this.$field.$type == "application/x-reference") {
			this.$filterField = this.field.$field.$item[this.field.$reference.$value.$itemProp];
			if (this.$filterField) {
				this.$filterField = this.record.$prototype.$properties[this.field.$reference.$value.$prop];
				if (this.$filterField) {
					this.$filterCode = this.field.$reference.$value.$prop;
				}
			}
			if (!this.$filterField) {
				this.$filterField = this.field.$field.$item.$properties[this.field.$reference.$value.$prop];
				this.$filterCode = this.$bind + "." + this.field.$reference.$value.$prop;
			}
		}
	},
	render: function(record, layoutSlot, $field, $bind) {
		this.$field = $field;
		this.record = record;
		this.$filterCode = this.$bind = $bind;
		this.operator = "none";
		this.layoutSlot = layoutSlot;
		this.loadField(this.$field, this.$bind);

		// update filter cell if where params in url (issue #3022)
		if (this.record.list.$urlParams.where) {
			var $filterBind = this.$bind;
			var exp = document.site.filterMaker.parseWhere(this.record.list.$urlParams.where, this.$bind);
			var $isRefField = false;
			// in case of reference field, appropriate $bind value is field.$reference.$value.$prop
			if ($field.$type == 'application/x-reference') {
				$filterBind = this.field.$reference.$value.$prop;
				$isRefField = true;
			}
			if (exp && $filterBind == exp.children[0].value) {
				this.setOperator(exp.value.text);
				var value = document.site.filterMaker.unFormatValue(exp.value.text, exp.children[1].value);
				// update dataset and field data value
				if ($isRefField) {
					this.record.dataset[this.$bind] = {};
					this.record.dataset[this.$bind][$filterBind] = value;
				} else {
					this.record.dataset[this.$bind] = value;
				}
				this.field.setDataValue(value);
			}
		}
	},

	setOperator: function(operator) {
		this._operator.style.backgroundImage = "url('" + document.site.$item.$iconPath + "list/s-filter-" + (this.operator = operator) + ".png')";
	},
	getValue: function() {
		var field = this.record.boundFields[this.$bind][0];
		if (this.operator != "none") {
			this._ensureFilterField();
			return document.site.filterMaker.getFieldValue({
				field: field,
				$bind: this.$bind,
				$type: this.$field.$type,
				operator: this.operator,
				$filterCode: this.$filterCode,
				$filterField: this.$filterField
			});
		}
	},
	loadField: function($field, $bind) {
		this.field = this.record.page.loadNewItem(this.layoutSlot.appendChild(document.createElement("div")), {
			$bind: $bind || this.$bind,
			$isTopLabelAlignment: false,
			$isCellChild: true,
			$isFilterMode: true,
			$isMenusHiddeen: true,
			$inplace: true,
			$field: $field
		}, this.record);
		this.addOperator();
	},
	ensureEqual: function() {
		// force operator to equals
		if (this.operator === "none" && this.$field) {
			var defaultFilter;
			switch (this.$field.$type) {
				case "application/x-date":
				case "application/x-time":
				case "application/x-datetime":
					defaultFilter = "ge";
					break;
				case "application/x-string":
					defaultFilter = "like";
					break;
				default:
					defaultFilter = "eq";
					break;
			}
			this.setOperator(defaultFilter);
		}
	},
	ensureDefault: function() {
		if (this.$field) {
			var defaultFilter, value = this.record.boundFields[this.$bind][0].getInputValue();
			switch (this.$field.$type) {
				case "application/x-date":
				case "application/x-time":
				case "application/x-datetime":
					defaultFilter = (value !== "" && this.operator === "none") ? "ge" : (value === "" && this.operator !== "none") ? "none" : undefined;
					break;
				case "application/x-string":
					defaultFilter = (value !== "" && this.operator === "none") ? "like_s" : (value === "" && this.operator !== "none") ? "none" : undefined;
					break;
				case "application/x-integer":
					defaultFilter = (value !== "" && this.operator === "none") ? "ge" : (value === "" && this.operator !== "none") ? "none" : undefined;
					break;
				default:
					defaultFilter = (value !== "" && this.operator === "none") ? "ge" : (value === "" && this.operator !== "none") ? "none" : undefined;
					break;
			}
			if (defaultFilter)
				this.setOperator(defaultFilter);
		}
	},
	dispose: function() {
		if (this._operator) {
			$(this._operator).unbind();
		}
		if (this.$$list) {
			this.$$list.undelegate();
		}
		this.$filterField = this.record = this.field = this.layoutSlot = this._operator = this.$$list = this.$field = null;
	}
});
