// @sage/syracuse-lib/src/license/dispatcher.ts
"use strict";

import * as check from '../license/check';
import * as mock from '../load-balancer/mock';
import * as url from 'url';
import * as config from 'config';
import { globals } from 'streamline-runtime';
var syracuse;

export function dispatcher(config) {
	var routes = {
		ub(_, request, response) {
			response.writeHead(200, {
				"Content-Type": "text/plain"
			});
			console.log("Find used badges");
			try {
				var result = check.findUsedBadges(_, "desktop");
				response.end(JSON.stringify(result));
			} catch (e) {
				console.log("Error " + e + " " + e.stack);
				response.end("Error during find used badges: " + e);
			}
		},

		test(_, request, response) {
			response.writeHead(200, {
				"Content-Type": "text/plain"
			});
			// This method is used by cloud platform to test healthcheck of the syracuse component.
			response.end('OK');
		},
		count(_, request, response) {
			response.writeHead(200, {
				"Content-Type": "text/plain"
			});
			var r = /c=(\d+)/.exec(request.url);
			var res = 0;
			if (r) {
				res = require('../license/check').step(_, r[1], {
					"x3ServerHost": "aws-x3-indv7",
					"x3ServerPort": 17100,
					"x3Solution": "X3V7",
					"x3Folder": "X3TESTV7",
					"locale": "en-US",
					"userName": "admin",
					"product": {
						"code": "1",
						"version": "7"
					}
				});

			}
			response.end('OK ' + r[1] + " " + res);

		},
		check(_, request, response) {
			console.log("LIC CHECK");
			check.checkNamed(_); // no instance here in order to avoid infinite loop!
			console.log("LIC CHECK");
			response.writeHead(200, {
				"Content-Type": "text/plain"
			});
			response.end('OK');
		},
		update(_, request, response) {
			var content = request.readAll(_).toString("utf8");
			console.log("Update request");
			var updateResult = check.updateLicense(content, _);
			response.writeHead(200, {
				"Content-Type": "text/plain"
			});
			response.end(updateResult ? 'Valid license' : 'No valid license');
		},
		set(_, request, response) {
			var content = request.readAll(_);
			if (content) content = content.toString("utf8");
			else {
				response.writeHead(400, {
					"Content-Type": "text/plain"
				});
				response.end("No content");
				return;
			}
			var diagnoses = [];
			var extra = {};
			try {
				check.licenseChange(content, diagnoses, _, extra);
			} catch (e) {
				response.writeHead(400, {
					"Content-Type": "text/plain"
				});
				response.end("Error: " + e);
				return;
			}
			response.writeHead(200, {
				"Content-Type": "text/plain"
			});
			response.end("Call: " + JSON.stringify(diagnoses) + "\nPropagation: " + (extra.error || extra.answer));
			return;
		}
	};
	return function(_, request, response) {
		// block special functions when not authorized
		if (request.fromNanny || (request._request && request._request.fromNanny) ||  
			(config.system && config.system.unsafeForceNanny)) {
			if (!syracuse) {
				syracuse = require('syracuse-main/lib/syracuse');
			}
			var multiTenant = config.hosting && config.hosting.multiTenant;
			if (multiTenant) {
				var r = /\?(?:.*\&)*tenantId=(\w+)/.exec(request.url);
				var tenant;
				if (r) tenant = r[1];
				if (!syracuse.initializedTenant(tenant)) {
					response.writeHead("200", {});
					return response.end("Ignored");
				} else {
					globals.context.tenantId = tenant;
				}
			}
			var p = request.url.split('/')[2];
			var index = p.indexOf('?');
			if (index >= 0) p = p.substr(0, index);
			var route = routes[p];
			if (!route) throw new Error("bad url: " + request.url);
			return route(_, request, response);
		} else {
			response.writeHead("404", {});
			return response.end("Resource not found.");
		}
	};
};
