// msoffice/lib/excel/contract/representations/excelDatasource.js
"use strict";

exports.representation = {
    $entityName: "excelDatasource",
    $query: {
        $layout: {
            "$items": [{
                "$bind": "$resources",
                "$format": "cards",
                "$cardsByRowCount": 2,
                "$layout": {
                    "$layoutType": "stack",
                    "$items": [{
                        "$items": [{
                            "$bind": "dsName"
                        }, {
                            "$bind": "description"
                        }, {
                            "$bind": "filter"
                        }, {
                            "$bind": "orderBys"
                        }, {
                            "$layoutType": "row",
                            "$widths": "30,70",
                            "$items": [{
                                "$items": [{
                                    "$bind": "fetchAll"
                                }, {
                                    "$bind": "fetchLimit"
                                }]
                            }]
                        }]
                    }]
                }
            }]
        }
    },
    $details: {
        $layout: {
            $category: "section",
            $items: [{
                $layout: {
                    $items: [{
                        $bind: "title"
                    }, {
                        $bind: "entity"
                    }, {
                        $bind: "filter",
                    }, {
                        "$bind": "fetchAll",
                        "$isEditMode": true
                    }, {
                        "$bind": "fetchLimit",
                        "$isEditMode": true,
                        "$isTopLabelAlignment": false
                    }]
                }
            }]
        }
    },
    $edit: {
        $layout: {
            $items: [{
                $bind: "description",
                $isEditMode: false,
                $isTopAlignement: true
            }, {
                $bind: "title"
            }, {
                $bind: "filter"
            }, {
                $bind: "fetchAll"
            }, {
                $bind: "fetchLimit"
            }, {
                $bind: "orderBys"
            }]
        }
    }
};
