// @sage/syracuse-ui/lib/help/helpCenter.js
"use strict";
var LegalPage = require('@sage/syracuse-ui/lib/help/legalPage').LegalPage;
var _browserTab = {};
var _timeoutShortCut;

exports.openByShortCut = function(page, field) {
	if (syra_context.browser.isMSIE) {
		if (!_timeoutShortCut) {
			document.onhelp = function() {
				return (false);
			};
			window.onhelp = function() {
				return (false);
			};
		}
		_timeoutShortCut && clearTimeout(_timeoutShortCut);
		_timeoutShortCut = setTimeout(function() {
			_timeoutShortCut = null;
			document.onhelp = null;
			window.onhelp = null;
		}, 1000);
	}
	page.externalAdapter.onFieldClickPicker({
		field: field || page,
		pickerType: "help",
		doEvent: function() {
			exports.open(page, field);
		}
	});
};


exports.open = function(page, field, $url) {
	var winId = "sageerphelp";
	if (!$url) {
		var $links = page.$prototype.$links;
		$url = $links && $links.$help && $links.$help.$url;
		if ($url) {
			if (field) {
				$url = syra_expression.parse(field.articleParent, $url) + '#' + field.$item.$bind;
			} else {
				$url = syra_expression.parse(page, $url);
			}
		}
	}
	if ($url) {
		if (syra_context.browser.isMSIE) {
			// Troubles with IE... to improve!
			window.open($url, "_blank");
			return;
		}
		if (!_browserTab.winRef || _browserTab.winRef.closed == undefined || _browserTab.winRef.closed) {
			_browserTab.winRef = window.open($url, winId, "width=800,height=600,resizable,scrollbars,status");
		} else
		if (_browserTab.lastUrl != $url) {
			_browserTab.winRef = window.open($url, winId, "resizable=true,scrollbars=yes,status=yes");
			if (_browserTab.winRef && _browserTab.winRef.focus) {
				_browserTab.winRef.focus();
			}
		} else
		if (_browserTab.winRef && _browserTab.winRef.focus) {
			_browserTab.winRef.focus();
		}
		_browserTab.lastUrl = $url;
	}
};


exports.openLegalPage = function() {
	syra_ajax.get({
		url: syra_config.$legalPrototypeUrl,
		end: function(response) {
			if (response.ok) {
				var page = new LegalPage();
				page.load(response.data);
				syra_ajax.get({
					url: syra_config.$legalDataUrl,
					end: function(response) {
						if (response.ok) {
							page.applyChange(response.data);
							syra_over.renderPageModal(page, {
								authoringDisabled: true
							});
						} else {
							syra_alert.show(response.error.diagnoses, page);
						}
					}
				});
			} else {
				syra_alert.show(response.error.diagnoses);
			}
		}
	});

};

exports.dispose = function() {
	_browserTab = null;
};
