// @sage/syracuse-ui/lib/help/legalPage.js
"use strict";
var helpers = require('@sage/syracuse-core').helpers;
var DesktopPage = require("@sage/syracuse-ui/lib/page/desktopPage").DesktopPage;

function LegalPage() {}

exports.LegalPage = helpers.defineClass(LegalPage, DesktopPage, {
	load: function($prototype) {
		this.isMenuBarDisabled = true;
		this.layoutSlot = document.createElement("div");
		this.$prototype = $prototype;
		this.$prototype.$properties.$banner = {
			$type: "application/x-string"
		};

		this.$prototype.$links = this.$prototype.$links || {};
		this.$prototype.$links.$eula = {
			$title: syra_local.legalCopyright,
			$description: syra_local.legalEula,
			$target: "blank",
			$url: syra_config.$eulaUrl
		};
		this.$prototype.$links.$eulaDeps = {
			$title: syra_local.legalDepsTitle,
			$description: syra_local.legalDepsDescription,
			$target: "blank",
			$url: syra_config.$eulaDepsUrl
		};
		this.$prototype.$links.$banner = {
			$isDisabled: true,
			$title: "Legal banner"
		};

		var $skin = "s-site-legal";

		this.$item = {
			$category: "page",
			$css: $skin + "-header",
			$title: syra_local.legalPageTitle,

			$layout: {
				$items: [{
					$format: "cards",
					$id: "s-legal-page-list",
					$addMaximize: false,
					$isTitleHidden: true,
					$isPagerHidden: true,
					$isQuickDesignerDisabled: true,
					$isRowIndexVisible: false,
					$selectByRowIndex: false,
					$isMenuRecordHidden: true,
					$bind: "$resources",
					$isHoverDisabled: true,
					$skin: $skin + "-card",
					renderRecordContent: function(record) {
						var head = syra_dom.div("s-legal-item-head");
						syra_dom.text("s-legal-item-product", record.dataset.product, head);
						syra_dom.text("s-legal-item-version", record.dataset.productVersion, head);
						record.body.appendChild(head);
						var $rows = [
							["productIddn"],
							["serial", "valEnd"],
							["licenseeName", "resellerName"],
							["licenseeRegistrationNumber", "resellerRegistrationNumber"]
						];
						var tb = document.createElement("div");
						tb.className = "s-legal-item-body";
						for (var ii = 0, jj = $rows.length; ii < jj; ii++) {
							var $cells = $rows[ii];
							var row = document.createElement("div");
							row.className = "s-legal-item-row";
							for (var mm = 0, kk = $cells.length; mm < kk; mm++) {
								if (mm > 0) {
									syra_dom.div("s-legal-item-field-sep", row);
								}
								var $bind = $cells[mm];
								syra_dom.text("s-legal-item-field-title", record.$prototype.$properties[$bind].$title, row);
								record.page.addItem(syra_dom.div("s-legal-item-field-value", row), {
									$bind: $bind,
									$inplace: true
								}, record);

							}
							tb.appendChild(row);
						}
						record.body.appendChild(tb);
					},
					$layout: {
						$items: []
					}

				}]
			}
		};
		DesktopPage.prototype.load.call(this);
		var list = this.idMap["s-legal-page-list"];
		syra_dom.hide(list.topbar, true);
		syra_dom.hide(this.header, true);
		var footer = syra_dom.div("s-legal-footer", this.layoutContent.domItem);
		var warning = "Warning: This software is protected by the Intellectual Property Code, copyright laws and international treaties.";
		warning += " Any improper use of the software, including any partial or total reproduction or distribution of the software or any use beyond the rights acquired on the software is strictly prohibited.";
		warning += " Anyone not complying with these provisions shall be guilty of copyright infringement and will be liable to criminal penalties provided by the law.";
		syra_dom.text("s-site-legal-warning", warning, footer);

		this.addItem(footer, {
			$category: "link",
			$skin: "s-site-legal-copyright",
			$bind: "$eula"
		});
		this.addItem(footer, {
			$category: "link",
			$skin: "s-site-legal-copyright",
			$bind: "$eulaDeps"
		});
	},
	applyChange: function(newData) {
		if (newData) {
			if (newData.$resources && newData.$resources.length > 0) {
				for (var ii = 0, jj = newData.$resources.length; ii < jj; ii++) {
					newData.$resources[ii].productVersion = syra_local.legalProductVersionLabel + " " + newData.$resources[ii].productVersion;
				}
			}
			if (this.dataset && this.dataset.$copyright) {
				newData.$links = newData.$links || {};
				(newData.$links.$eula = newData.$links.$eula || {}).$title = this.dataset.$copyright;
				(newData.$links.$eulaDeps = newData.$links.$eulaDeps || {}).$title = syra_local.legalDepsTitle;
			}
			DesktopPage.prototype.applyChange.call(this, newData);
		}
	},
	onMenuClick: function(options) {
		if (options.menu.$bind == "$eula") {
			window.open(syra_config.$eulaUrl, "_blank");
			return false;
		}
		if (options.menu.$bind == "$eulaDeps") {
			window.open(syra_config.$eulaDepsUrl, "_blank");
			return false;
		}
		return true;
	}
});
