// @sage/syracuse-lib/src/collaboration/entities/page/landingPageVignette.ts
"use strict";

import { globals } from 'streamline-runtime';

export const entity = {
	$titleTemplate: "Vignettes",
	$valueTemplate: "{vignette.title} - {endpoint.description}",
	$properties: {
		bind: {
			$title: "Bind",
			$isHidden: true,
			defaultValue(_, instance) {
				return instance.$uuid;
			}
		},
		nature: {
			$title: "Nature",
			$compute(_, instance) {
				var ep = instance.endpoint(_);
				return ep && ep.nature(_);
			}
		}
	},
	$relations: {
		vignette: {
			$title: "Vignette",
			//			$type: "portlet",
			$type: "menuItem",
			$isMandatory: true,
			$lookupFilter: {
				//				type: "$page"
				linkType: {
					$ne: "$function"
				}
			}
		},
		endpoint: {
			$title: "Endpoint",
			$type: "endPoint",
			$isMandatory(_, instance) {
				return !instance._parent.useCurrentEndpoint(_);
			},
			$isDefined(_, instance) {
				return !instance._parent.useCurrentEndpoint(_);
			}
		}
	},
	$links(_, instance) {
		var lks = {
			$location: instance.getLocationLink(_)
		};
		return lks;
	},
	$functions: {
		$onDelete(_, instance) {
			globals.context.session && globals.context.session.resetCache && globals.context.session.resetCache("landingPage");
		},
		getLocationLink(_, baseUrlProp) {
			var res = {};
			var instance = this;
			//			var item = instance.vignette(_) && instance.vignette(_).pageItem(_);
			var item = instance.vignette(_);
			var ep = null;
			if (!instance._parent.useCurrentEndpoint(_)) ep = instance.endpoint(_);
			return item && item.getLink(_, ep, baseUrlProp || "{$selectedEpBaseUrl}");
		}
	},
	$events: {
		$afterSave: [

			function(_, instance, params) {
				globals.context.session && globals.context.session.resetCache && globals.context.session.resetCache("landingPage");
			}
		]
	}
};
