// msoffice/lib/word/wordInterface.js
"use strict"

var helpers = require('syracuse-core/lib/helpers');

var base64 = require("msoffice/lib/Base64").Base64;
var httpController = require("syracuse-ui/lib/common/controller/httpController");
var locale = require("syracuse-core/lib/locale");

function WordInterface() {
	var self = this;
}

exports.WordInterface = helpers.defineClass(WordInterface, null, {
	onLogin: function () {
		var self = this;
		self.loadPage( external.getSyracuseOfficeCustomData().getResourceUrl() );
	},
	loadPage: function(resourceUrl, name, onLoaded) {
		document.controller.loadRepresentation(
			null, 
			resourceUrl, 
			function(response) {
				var repr = response.$representation;
				repr.$article = (repr.$prototype || {}).$article || {};
				repr.$article.$category = "worksheet";
				repr.$prototype = repr.$prototype || {};
				// repr.$prototype.$datasource = datasource;
				// repr.$prototype.$datasourceName = name;
	
				var page = document.itemFactory.loadPage(
						{
						$representation: repr,
						$urlParts: response.$urlParts,
						$category: "worksheet",
						$autoFetch: false
						}
					);
				
				// TODO: Close browser here?
				external.closeDialogForm();				
			}, 
			function(error) { 
				alert(error); 
			}, 
			null);
	}
});
