// msoffice/lib/word/wordInterfaceImpl.js
"use strict"

// Used be wordInterface if word is opened

var helpers = require('syracuse-core/lib/helpers');
var msofficeHelpers = require('msoffice/lib/helpers');
var base64 = require("msoffice/lib/Base64").Base64;

function WordInterfaceImpl() {
}

exports.WordInterfaceImpl = helpers.defineClass(WordInterfaceImpl, null, {

	getResourceUrl: function () {
		return external.getSyracuseOfficeCustomData().getResourceUrl();
	},
	attachResourceUrl: function(url) {
		external.getSyracuseOfficeCustomData().setResourceUrl(url);
	},
	closeDialogForm: function() {
		external.closeBrowserDialog();
	},
	createDatasource: function(mailMergeData, url) {
		external.createDatasource(JSON.stringify(mailMergeData));
	},
	getDocumentContent: function() {
        return base64.raw_decode(external.GetDocumentContent());
	},
	getDocumentTitle: function() {
		return external.getSyracuseOfficeCustomData().getDocumentTitle();
	},
	setDocumentTitle: function(value) {
		external.getSyracuseOfficeCustomData().setDocumentTitle(value);
	},
	getDocumentUrl: function() {
		return external.getSyracuseOfficeCustomData().getDocumentUrl();
	},
	setDocumentUrl: function(value) {
		external.getSyracuseOfficeCustomData().setDocumentUrl(value);
	}	
});
