// msoffice/lib/word/wordInterfaceImpl.js
"use strict"

// Used be wordInterface if word is opened

var helpers = require('syracuse-core/lib/helpers');
var msofficeHelpers = require('msoffice/lib/helpers');
var base64 = require("msoffice/lib/Base64").Base64;

function WordInterfaceImpl() {
}

exports.WordInterfaceImpl = helpers.defineClass(WordInterfaceImpl, null, {

	getResourceUrl: function () {
		return external.getSyracuseOfficeCustomData().getResourceUrl();
	},
	attachResourceUrl: function(url) {
		external.getSyracuseOfficeCustomData().setResourceUrl(url);
	},
	createDatasource: function(mailMergeData, url) {
		external.createDatasource(JSON.stringify(mailMergeData));
	},
	createWordTemplate: function(layoutBoxes, data, refreshOnly) {
		var jsonData = JSON.stringify({ "layout": layoutBoxes , "data": data, "refreshOnly": refreshOnly});
		external.createWordTemplate(jsonData);	// add data to the template to be able to generate a preview
	},
	populateWordTemplate: function(data) {
		external.populateWordTemplate(JSON.stringify({ "data": data}));
	},
	// Used for saving doc to syracuse
	getDocumentContent: function() {
        return base64.raw_decode(external.GetDocumentContent());
	},
	getDocumentTitle: function() {
		return external.getSyracuseOfficeCustomData().getDocumentTitle();
	},
	setDocumentTitle: function(value) {
		external.getSyracuseOfficeCustomData().setDocumentTitle(value);
		external.getSyracuseOfficeCustomData().writeDictionaryToDocument();
	},
	getDocumentUrl: function() {
		return external.getSyracuseOfficeCustomData().getDocumentUrl();
	},
	setDocumentUrl: function(value) {
		external.getSyracuseOfficeCustomData().setDocumentUrl(value);
		external.getSyracuseOfficeCustomData().writeDictionaryToDocument();
	},
	notifySaveDocumentDone: function(value) {
		// Just notify user
		external.NotifySaveDocumentDone();	
	},
	getSyracuseDocumentType: function() {
		// return "word-mailmerge";
		// return "word-report";
		// return "word-report-tpl";		
		return external.getSyracuseDocumentType();
	},
	getAddinVersion: function() {
		return external.getAddinVersion();
	}
});
