// msoffice/lib/word/wordInterfaceDummy.js
"use strict";

// Used be wordInterface if word is not opened (testing in browser)

var helpers = require('syracuse-core/lib/helpers');
var msofficeHelpers = require('msoffice/lib/helpers');

function WordInterfaceDummy() {
	var self = this;
}
exports.WordInterfaceDummy = helpers.defineClass(WordInterfaceDummy, null, {
	
	getResourceUrl: function () {
		var val = this._getBaseURIParam("url");
		return val;
	},
	attachResourceUrl: function(url) {
		console.log("Attach resource: " + url);
	},
	createDatasource: function(mailMergeData, url) {
		console.log("Create datasource: " + url + "\n" + JSON.stringify(mailMergeData));
	},
	createWordTemplate: function(layoutBoxes) {
		alert("Extracted layout:\n" + JSON.stringify(layoutBoxes));
	},
	populateWordTemplate: function(data) {
		alert("Data:\n" + JSON.stringify(data));
	},
	downloadV6Document: function() {
	},
	getDocumentContent: function() {
		console.log("Read document content");
        return "Dummy content";
	},
	getDocumentTitle: function() {
	},
	setDocumentTitle: function(value) {
	},
	getDocumentUrl: function() {
	},
	setDocumentUrl: function(value) {
	},
	getDocumentRepresentation: function() {
		return "noentity.$nofacet";
	},
	getDocumentLocale: function() {
		return "de-DE";
	},
	getOriginalFileName: function() {
		return "doc-de";
	},
	notifySaveDocumentDone: function(value) {
	},
	getSyracuseDocumentType: function() {
		var val = this._getBaseURIParam("mode");
		return val;
	},
	getAddinVersion: function() {
	},
	_getBaseURIParam: function(param) {
		var params = document.baseURI.split("?")[1].split("&");
		var val;
		params.forEach(function(p) {
			var idx = p.indexOf(param + "=");
			if (idx > -1)
			{
				val = p.substring(idx + param.length + 1);
				val = decodeURIComponent(val);
			}
		});
		return val;
	}
});
