// syracuse-ui/lib/tests/misc/formulaTest.js
"use strict";
var helpers = require("syracuse-core/lib/helpers");
var TestPage = require("syracuse-ui/lib/tests/testPage").TestPage;
var formula = require("syracuse-ui/lib/site/aside/formula");

function FormulaTest() {}

exports.FormulaTest = helpers.defineClass(FormulaTest, TestPage, {
	_initiliazeTestPage: function() {
		var $formulaPrototype = formula.getFormulaPage();
		syra_site.deltaManager.applyObjectDelta(this, this.$prototype.$properties, $formulaPrototype.$properties, true);
		syra_site.deltaManager.applyObjectDelta(this, this.$prototype.$links, $formulaPrototype.$links, true);
		this.$prototype.$facet = "$edit";
		this.$prototype.$actions.$openDialog = {
			$title: "openDialog"
		};
		this.$item.$layout = {
			$items: [{
					$category: "link",
					$bind: "$openDialog"
				},
				$formulaPrototype.$article.$layout
			]
		};
	},
	loadBox: function() {
		TestPage.prototype.loadBox.call(this);
		this.fillPage(this);
	},
	notifyActionChange: function(menuItem, target, value) {
		if (value.$openDialog) {
			var self = this;
			syra_site.dialogManager.openModal(self, {
				$itemPage: {
					$facet: "edit",
					$representation: {
						$prototype: formula.getFormulaPage()
					},
					onAfterLoadPage: function(page) {
						self.fillPage(page);
					}
				}
			});
		}
	},
	fillPage: function(page) {
		page.applyChange({
			"fl": [{
				"$uuid": 0,
				"$serverIndex": 0,
				"$rawV": {
					"fl1": {},
					"fl2": {},
					"fl3": {}
				},
				"$key": "122~ZZ1",
				"fl2": "ZZ1",
				"fl1": "",
				"fl3": "Nouvelle table/vue",
				"fl4": 2,
				"extends0": "",
				"extends1": "ZZ1",
				"extends2": "Nouvelle table/vue",
				"extends3": 5,
				"extends4": "Aide de la table/vue",
				"extends5": "ZZ1"
			}, {
				"$uuid": 1,
				"$serverIndex": 1,
				"$rawV": {
					"fl1": {},
					"fl2": {},
					"fl3": {}
				},
				"$key": "122~ZZ2",
				"fl2": "ZZ2",
				"fl1": "",
				"fl3": "Nouvel écran",
				"fl4": 2,
				"extends0": "",
				"extends1": "ZZ2",
				"extends2": "Nouvel écran",
				"extends3": 5,
				"extends4": "Aide de l'écran",
				"extends5": "ZZ2"
			}, {
				"$uuid": 2,
				"$serverIndex": 2,
				"$rawV": {
					"fl1": {},
					"fl2": {},
					"fl3": {}
				},
				"$key": "122~VARIABLE",
				"fl2": "VARIABLE",
				"fl1": "",
				"fl3": "Variables globales",
				"fl4": 2,
				"extends0": "",
				"extends1": "VARIABLE",
				"extends2": "Variables globales",
				"extends3": 1,
				"extends4": "Aide de la table/vue",
				"extends5": "Variable"
			}, {
				"$uuid": 3,
				"$serverIndex": 3,
				"$rawV": {
					"fl1": {},
					"fl2": {},
					"fl3": {}
				},
				"$key": "122~VARIABLE.GDEFSOC",
				"fl2": "VARIABLE.GDEFSOC",
				"fl1": "VARIABLE",
				"fl3": "GDEFSOC Société par défaut",
				"fl4": 1,
				"extends0": "VARIABLE",
				"extends1": "VARIABLE.GDEFSOC",
				"extends2": "GDEFSOC Société par défaut",
				"extends3": 2,
				"extends4": "Aide du champ",
				"extends5": "GDEFSOC"
			}, {
				"$uuid": 4,
				"$serverIndex": 4,
				"$rawV": {
					"fl1": {},
					"fl2": {},
					"fl3": {}
				},
				"$key": "122~VARIABLE.GERRBATCH",
				"fl2": "VARIABLE.GERRBATCH",
				"fl1": "VARIABLE",
				"fl3": "GERRBATCH Nb d'erreurs dans la tâche",
				"fl4": 1,
				"extends0": "VARIABLE",
				"extends1": "VARIABLE.GERRBATCH",
				"extends2": "GERRBATCH Nb d'erreurs dans la tâche",
				"extends3": 2,
				"extends4": "Aide du champ",
				"extends5": "GERRBATCH"
			}, {
				"$uuid": 5,
				"$serverIndex": 5,
				"$rawV": {
					"fl1": {},
					"fl2": {},
					"fl3": {}
				},
				"$key": "122~VARIABLE.GERRTRACE",
				"fl2": "VARIABLE.GERRTRACE",
				"fl1": "VARIABLE",
				"fl3": "GERRTRACE Nb d'erreurs dans la trace",
				"fl4": 1,
				"extends0": "VARIABLE",
				"extends1": "VARIABLE.GERRTRACE",
				"extends2": "GERRTRACE Nb d'erreurs dans la trace",
				"extends3": 2,
				"extends4": "Aide du champ",
				"extends5": "GERRTRACE"
			}, {
				"$uuid": 6,
				"$serverIndex": 6,
				"$rawV": {
					"fl1": {},
					"fl2": {},
					"fl3": {}
				},
				"$key": "122~VARIABLE.GFCYDEF",
				"fl2": "VARIABLE.GFCYDEF",
				"fl1": "VARIABLE",
				"fl3": "GFCYDEF Site par défaut",
				"fl4": 1,
				"extends0": "VARIABLE",
				"extends1": "VARIABLE.GFCYDEF",
				"extends2": "GFCYDEF Site par défaut",
				"extends3": 2,
				"extends4": "Aide du champ",
				"extends5": "GFCYDEF"
			}, {
				"$uuid": 7,
				"$serverIndex": 7,
				"$rawV": {
					"fl1": {},
					"fl2": {},
					"fl3": {}
				},
				"$key": "122~VARIABLE.GIMPORT",
				"fl2": "VARIABLE.GIMPORT",
				"fl1": "VARIABLE",
				"fl3": "GIMPORT Import en cours",
				"fl4": 1,
				"extends0": "VARIABLE",
				"extends1": "VARIABLE.GIMPORT",
				"extends2": "GIMPORT Import en cours",
				"extends3": 2,
				"extends4": "Aide du champ",
				"extends5": "GIMPORT"
			}, {
				"$uuid": 8,
				"$serverIndex": 8,
				"$rawV": {
					"fl1": {},
					"fl2": {},
					"fl3": {}
				},
				"$key": "122~VARIABLE.GLANDEF",
				"fl2": "VARIABLE.GLANDEF",
				"fl1": "VARIABLE",
				"fl3": "GLANDEF Langue par défaut",
				"fl4": 1,
				"extends0": "VARIABLE",
				"extends1": "VARIABLE.GLANDEF",
				"extends2": "GLANDEF Langue par défaut",
				"extends3": 2,
				"extends4": "Aide du champ",
				"extends5": "GLANDEF"
			}, {
				"$uuid": 9,
				"$serverIndex": 9,
				"$rawV": {
					"fl1": {},
					"fl2": {},
					"fl3": {}
				},
				"$key": "122~VARIABLE.GLANGUE",
				"fl2": "VARIABLE.GLANGUE",
				"fl1": "VARIABLE",
				"fl3": "GLANGUE Langue courante",
				"fl4": 1,
				"extends0": "VARIABLE",
				"extends1": "VARIABLE.GLANGUE",
				"extends2": "GLANGUE Langue courante",
				"extends3": 2,
				"extends4": "Aide du champ",
				"extends5": "GLANGUE"
			}, {
				"$uuid": 10,
				"$serverIndex": 10,
				"$rawV": {
					"fl1": {},
					"fl2": {},
					"fl3": {}
				},
				"$key": "122~VARIABLE.GMESSBATCH",
				"fl2": "VARIABLE.GMESSBATCH",
				"fl1": "VARIABLE",
				"fl3": "GMESSBATCH Message d'erreur dans la tâche",
				"fl4": 1,
				"extends0": "VARIABLE",
				"extends1": "VARIABLE.GMESSBATCH",
				"extends2": "GMESSBATCH Message d'erreur dans la tâche",
				"extends3": 2,
				"extends4": "Aide du champ",
				"extends5": "GMESSBATCH"
			}, {
				"$uuid": 11,
				"$serverIndex": 11,
				"$rawV": {
					"fl1": {},
					"fl2": {},
					"fl3": {}
				},
				"$key": "122~VARIABLE.GNOMUSER",
				"fl2": "VARIABLE.GNOMUSER",
				"fl1": "VARIABLE",
				"fl3": "GNOMUSER Nom utilisateur",
				"fl4": 1,
				"extends0": "VARIABLE",
				"extends1": "VARIABLE.GNOMUSER",
				"extends2": "GNOMUSER Nom utilisateur",
				"extends3": 2,
				"extends4": "Aide du champ",
				"extends5": "GNOMUSER"
			}, {
				"$uuid": 12,
				"$serverIndex": 12,
				"$rawV": {
					"fl1": {},
					"fl2": {},
					"fl3": {}
				},
				"$key": "122~VARIABLE.GOPTCANA",
				"fl2": "VARIABLE.GOPTCANA",
				"fl1": "VARIABLE",
				"fl3": "GOPTCANA Nombre d'axes analytiques",
				"fl4": 1,
				"extends0": "VARIABLE",
				"extends1": "VARIABLE.GOPTCANA",
				"extends2": "GOPTCANA Nombre d'axes analytiques",
				"extends3": 2,
				"extends4": "Aide du champ",
				"extends5": "GOPTCANA"
			}, {
				"$uuid": 13,
				"$serverIndex": 13,
				"$rawV": {
					"fl1": {},
					"fl2": {},
					"fl3": {}
				},
				"$key": "122~VARIABLE.GSYSCUR",
				"fl2": "VARIABLE.GSYSCUR",
				"fl1": "VARIABLE",
				"fl3": "GSYSCUR Devise de reporting",
				"fl4": 1,
				"extends0": "VARIABLE",
				"extends1": "VARIABLE.GSYSCUR",
				"extends2": "GSYSCUR Devise de reporting",
				"extends3": 2,
				"extends4": "Aide du champ",
				"extends5": "GSYSCUR"
			}, {
				"$uuid": 14,
				"$serverIndex": 14,
				"$rawV": {
					"fl1": {},
					"fl2": {},
					"fl3": {}
				},
				"$key": "122~VARIABLE.GUSER",
				"fl2": "VARIABLE.GUSER",
				"fl1": "VARIABLE",
				"fl3": "GUSER Code utilisateur",
				"fl4": 1,
				"extends0": "VARIABLE",
				"extends1": "VARIABLE.GUSER",
				"extends2": "GUSER Code utilisateur",
				"extends3": 2,
				"extends4": "Aide du champ",
				"extends5": "GUSER"
			}],
			"$isPartialDelta": false
		});
		page.applyChange({
			"operator": [{
				"$uuid": 0,
				"$serverIndex": 0,
				"$rawV": {
					"operator1": {},
					"operator2": {},
					"operator3": {}
				},
				"$key": "122~ONUM",
				"operator2": "ONUM",
				"operator1": "",
				"operator3": "Numérique",
				"operator4": 2,
				"extends0": "",
				"extends1": "ONUM",
				"extends2": "Numérique",
				"extends3": 1,
				"extends4": "",
				"extends5": ""
			}, {
				"$uuid": 1,
				"$serverIndex": 1,
				"$rawV": {
					"operator1": {},
					"operator2": {},
					"operator3": {}
				},
				"$key": "122~OPLUS",
				"operator2": "OPLUS",
				"operator1": "ONUM",
				"operator3": "Addition  '+'",
				"operator4": 2,
				"extends0": "ONUM",
				"extends1": "OPLUS",
				"extends2": "Addition  '+'",
				"extends3": 1,
				"extends4": "",
				"extends5": " + %"
			}, {
				"$uuid": 2,
				"$serverIndex": 2,
				"$rawV": {
					"operator1": {},
					"operator2": {},
					"operator3": {}
				},
				"$key": "122~OMOINS",
				"operator2": "OMOINS",
				"operator1": "ONUM",
				"operator3": "Soustraction  '-'",
				"operator4": 2,
				"extends0": "ONUM",
				"extends1": "OMOINS",
				"extends2": "Soustraction  '-'",
				"extends3": 1,
				"extends4": "",
				"extends5": " - %"
			}, {
				"$uuid": 3,
				"$serverIndex": 3,
				"$rawV": {
					"operator1": {},
					"operator2": {},
					"operator3": {}
				},
				"$key": "122~OMULTIPLICATION",
				"operator2": "OMULTIPLICATION",
				"operator1": "ONUM",
				"operator3": "Multiplication  '*'",
				"operator4": 2,
				"extends0": "ONUM",
				"extends1": "OMULTIPLICATION",
				"extends2": "Multiplication  '*'",
				"extends3": 1,
				"extends4": "",
				"extends5": " * %"
			}, {
				"$uuid": 4,
				"$serverIndex": 4,
				"$rawV": {
					"operator1": {},
					"operator2": {},
					"operator3": {}
				},
				"$key": "122~ODIVISION",
				"operator2": "ODIVISION",
				"operator1": "ONUM",
				"operator3": "Division  '/'",
				"operator4": 2,
				"extends0": "ONUM",
				"extends1": "ODIVISION",
				"extends2": "Division  '/'",
				"extends3": 1,
				"extends4": "",
				"extends5": " / %"
			}, {
				"$uuid": 5,
				"$serverIndex": 5,
				"$rawV": {
					"operator1": {},
					"operator2": {},
					"operator3": {}
				},
				"$key": "122~OPUISSANCE",
				"operator2": "OPUISSANCE",
				"operator1": "ONUM",
				"operator3": "Puissance  '^'",
				"operator4": 2,
				"extends0": "ONUM",
				"extends1": "OPUISSANCE",
				"extends2": "Puissance  '^'",
				"extends3": 1,
				"extends4": "",
				"extends5": " ^ %"
			}, {
				"$uuid": 6,
				"$serverIndex": 6,
				"$rawV": {
					"operator1": {},
					"operator2": {},
					"operator3": {}
				},
				"$key": "122~OCOMP",
				"operator2": "OCOMP",
				"operator1": "",
				"operator3": "Comparaison",
				"operator4": 2,
				"extends0": "",
				"extends1": "OCOMP",
				"extends2": "Comparaison",
				"extends3": 2,
				"extends4": "",
				"extends5": ""
			}, {
				"$uuid": 7,
				"$serverIndex": 7,
				"$rawV": {
					"operator1": {},
					"operator2": {},
					"operator3": {}
				},
				"$key": "122~OEGAL",
				"operator2": "OEGAL",
				"operator1": "OCOMP",
				"operator3": "Egal  '='",
				"operator4": 2,
				"extends0": "OCOMP",
				"extends1": "OEGAL",
				"extends2": "Egal  '='",
				"extends3": 2,
				"extends4": "",
				"extends5": " = %"
			}, {
				"$uuid": 8,
				"$serverIndex": 8,
				"$rawV": {
					"operator1": {},
					"operator2": {},
					"operator3": {}
				},
				"$key": "122~ODIFFERENT",
				"operator2": "ODIFFERENT",
				"operator1": "OCOMP",
				"operator3": "Différent  '<>'",
				"operator4": 2,
				"extends0": "OCOMP",
				"extends1": "ODIFFERENT",
				"extends2": "Différent  '<>'",
				"extends3": 2,
				"extends4": "",
				"extends5": " <> %"
			}, {
				"$uuid": 9,
				"$serverIndex": 9,
				"$rawV": {
					"operator1": {},
					"operator2": {},
					"operator3": {}
				},
				"$key": "122~OINFERIEUR",
				"operator2": "OINFERIEUR",
				"operator1": "OCOMP",
				"operator3": "Inférieur  '<'",
				"operator4": 2,
				"extends0": "OCOMP",
				"extends1": "OINFERIEUR",
				"extends2": "Inférieur  '<'",
				"extends3": 2,
				"extends4": "",
				"extends5": " < %"
			}, {
				"$uuid": 10,
				"$serverIndex": 10,
				"$rawV": {
					"operator1": {},
					"operator2": {},
					"operator3": {}
				},
				"$key": "122~OINFERIEUREGAL",
				"operator2": "OINFERIEUREGAL",
				"operator1": "OCOMP",
				"operator3": "Inférieur ou égal'<='",
				"operator4": 2,
				"extends0": "OCOMP",
				"extends1": "OINFERIEUREGAL",
				"extends2": "Inférieur ou égal'<='",
				"extends3": 2,
				"extends4": "",
				"extends5": " <= %"
			}, {
				"$uuid": 11,
				"$serverIndex": 11,
				"$rawV": {
					"operator1": {},
					"operator2": {},
					"operator3": {}
				},
				"$key": "122~OSUPERIEUR",
				"operator2": "OSUPERIEUR",
				"operator1": "OCOMP",
				"operator3": "Supérieur  '>'",
				"operator4": 2,
				"extends0": "OCOMP",
				"extends1": "OSUPERIEUR",
				"extends2": "Supérieur  '>'",
				"extends3": 2,
				"extends4": "",
				"extends5": " > %"
			}, {
				"$uuid": 12,
				"$serverIndex": 12,
				"$rawV": {
					"operator1": {},
					"operator2": {},
					"operator3": {}
				},
				"$key": "122~OSUPERIEUREGAL",
				"operator2": "OSUPERIEUREGAL",
				"operator1": "OCOMP",
				"operator3": "Supérieur ou égal'>='",
				"operator4": 2,
				"extends0": "OCOMP",
				"extends1": "OSUPERIEUREGAL",
				"extends2": "Supérieur ou égal'>='",
				"extends3": 2,
				"extends4": "",
				"extends5": " >= %"
			}, {
				"$uuid": 13,
				"$serverIndex": 13,
				"$rawV": {
					"operator1": {},
					"operator2": {},
					"operator3": {}
				},
				"$key": "122~OLOG",
				"operator2": "OLOG",
				"operator1": "",
				"operator3": "Logique",
				"operator4": 2,
				"extends0": "",
				"extends1": "OLOG",
				"extends2": "Logique",
				"extends3": 3,
				"extends4": "",
				"extends5": ""
			}, {
				"$uuid": 14,
				"$serverIndex": 14,
				"$rawV": {
					"operator1": {},
					"operator2": {},
					"operator3": {}
				},
				"$key": "122~OET",
				"operator2": "OET",
				"operator1": "OLOG",
				"operator3": "Et  '&'",
				"operator4": 2,
				"extends0": "OLOG",
				"extends1": "OET",
				"extends2": "Et  '&'",
				"extends3": 3,
				"extends4": "",
				"extends5": " & %"
			}, {
				"$uuid": 15,
				"$serverIndex": 15,
				"$rawV": {
					"operator1": {},
					"operator2": {},
					"operator3": {}
				},
				"$key": "122~OOU",
				"operator2": "OOU",
				"operator1": "OLOG",
				"operator3": "Ou  '|'",
				"operator4": 2,
				"extends0": "OLOG",
				"extends1": "OOU",
				"extends2": "Ou  '|'",
				"extends3": 3,
				"extends4": "",
				"extends5": " | %"
			}, {
				"$uuid": 16,
				"$serverIndex": 16,
				"$rawV": {
					"operator1": {},
					"operator2": {},
					"operator3": {}
				},
				"$key": "122~ONON",
				"operator2": "ONON",
				"operator1": "OLOG",
				"operator3": "Non  '!'",
				"operator4": 2,
				"extends0": "OLOG",
				"extends1": "ONON",
				"extends2": "Non  '!'",
				"extends3": 3,
				"extends4": "",
				"extends5": " ! %"
			}, {
				"$uuid": 17,
				"$serverIndex": 17,
				"$rawV": {
					"operator1": {},
					"operator2": {},
					"operator3": {}
				},
				"$key": "122~OOUEXCLUSIF",
				"operator2": "OOUEXCLUSIF",
				"operator1": "OLOG",
				"operator3": "Ou exclusif '?'",
				"operator4": 2,
				"extends0": "OLOG",
				"extends1": "OOUEXCLUSIF",
				"extends2": "Ou exclusif '?'",
				"extends3": 3,
				"extends4": "",
				"extends5": " ? %"
			}],
			"$isPartialDelta": false
		});
		page.applyChange({
			"fc": [{
				"$uuid": 0,
				"$serverIndex": 0,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FARITH",
				"fc2": "FARITH",
				"fc1": "",
				"fc3": "Arithmétique",
				"fc4": 1,
				"extends0": "",
				"extends1": "FARITH",
				"extends2": "Arithmétique",
				"extends3": 1,
				"extends4": "",
				"extends5": ""
			}, {
				"$uuid": 1,
				"$serverIndex": 1,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FABS",
				"fc2": "FABS",
				"fc1": "FARITH",
				"fc3": "Valeur absolue  'abs(nbr)'",
				"fc4": 1,
				"extends0": "FARITH",
				"extends1": "FABS",
				"extends2": "Valeur absolue  'abs(nbr)'",
				"extends3": 1,
				"extends4": "Abs",
				"extends5": "abs(%)"
			}, {
				"$uuid": 2,
				"$serverIndex": 2,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FAR2",
				"fc2": "FAR2",
				"fc1": "FARITH",
				"fc3": "Arrondi 2 décimales 'ar2(nbr)'",
				"fc4": 1,
				"extends0": "FARITH",
				"extends1": "FAR2",
				"extends2": "Arrondi 2 décimales 'ar2(nbr)'",
				"extends3": 1,
				"extends4": "Ar2",
				"extends5": "ar2(%)"
			}, {
				"$uuid": 3,
				"$serverIndex": 3,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FARR",
				"fc2": "FARR",
				"fc1": "FARITH",
				"fc3": "Arrondi  'arr(nbr,nbr)'",
				"fc4": 1,
				"extends0": "FARITH",
				"extends1": "FARR",
				"extends2": "Arrondi  'arr(nbr,nbr)'",
				"extends3": 1,
				"extends4": "Ar2",
				"extends5": "arr(%, )"
			}, {
				"$uuid": 4,
				"$serverIndex": 4,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FAVG",
				"fc2": "FAVG",
				"fc1": "FARITH",
				"fc3": "Moyenne  'avg(nbr)'",
				"fc4": 1,
				"extends0": "FARITH",
				"extends1": "FAVG",
				"extends2": "Moyenne  'avg(nbr)'",
				"extends3": 1,
				"extends4": "Avg",
				"extends5": "avg(%)"
			}, {
				"$uuid": 5,
				"$serverIndex": 5,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FFIX",
				"fc2": "FFIX",
				"fc1": "FARITH",
				"fc3": "Troncature  'fix(nbr)'",
				"fc4": 1,
				"extends0": "FARITH",
				"extends1": "FFIX",
				"extends2": "Troncature  'fix(nbr)'",
				"extends3": 1,
				"extends4": "Fix",
				"extends5": "fix(%)"
			}, {
				"$uuid": 6,
				"$serverIndex": 6,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FINT",
				"fc2": "FINT",
				"fc1": "FARITH",
				"fc3": "Partie entière 'int(nbr)'",
				"fc4": 1,
				"extends0": "FARITH",
				"extends1": "FINT",
				"extends2": "Partie entière 'int(nbr)'",
				"extends3": 1,
				"extends4": "Int",
				"extends5": "int(%)"
			}, {
				"$uuid": 7,
				"$serverIndex": 7,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FMAX",
				"fc2": "FMAX",
				"fc1": "FARITH",
				"fc3": "Maximum  'max(liste)'",
				"fc4": 1,
				"extends0": "FARITH",
				"extends1": "FMAX",
				"extends2": "Maximum  'max(liste)'",
				"extends3": 1,
				"extends4": "Max",
				"extends5": "max(%, )"
			}, {
				"$uuid": 8,
				"$serverIndex": 8,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FMIN",
				"fc2": "FMIN",
				"fc1": "FARITH",
				"fc3": "Minimum  'min(liste)'",
				"fc4": 1,
				"extends0": "FARITH",
				"extends1": "FMIN",
				"extends2": "Minimum  'min(liste)'",
				"extends3": 1,
				"extends4": "Min",
				"extends5": "min(%, )"
			}, {
				"$uuid": 9,
				"$serverIndex": 9,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FMOD",
				"fc2": "FMOD",
				"fc1": "FARITH",
				"fc3": "Reste de division entière  'mod(nbr,nbr)'",
				"fc4": 1,
				"extends0": "FARITH",
				"extends1": "FMOD",
				"extends2": "Reste de division entière  'mod(nbr,nbr)'",
				"extends3": 1,
				"extends4": "Mod",
				"extends5": "mod(%, )"
			}, {
				"$uuid": 10,
				"$serverIndex": 10,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FRND",
				"fc2": "FRND",
				"fc1": "FARITH",
				"fc3": "Nombre au hasard 'rnd(nbr)'",
				"fc4": 1,
				"extends0": "FARITH",
				"extends1": "FRND",
				"extends2": "Nombre au hasard 'rnd(nbr)'",
				"extends3": 1,
				"extends4": "Rnd",
				"extends5": "rnd(%)"
			}, {
				"$uuid": 11,
				"$serverIndex": 11,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FSGN",
				"fc2": "FSGN",
				"fc1": "FARITH",
				"fc3": "Signe  'sgn(nbr)'",
				"fc4": 1,
				"extends0": "FARITH",
				"extends1": "FSGN",
				"extends2": "Signe  'sgn(nbr)'",
				"extends3": 1,
				"extends4": "Sgn",
				"extends5": "sgn(%)"
			}, {
				"$uuid": 12,
				"$serverIndex": 12,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FSQR",
				"fc2": "FSQR",
				"fc1": "FARITH",
				"fc3": "Racine carrée  'sqr(nbr)'",
				"fc4": 1,
				"extends0": "FARITH",
				"extends1": "FSQR",
				"extends2": "Racine carrée  'sqr(nbr)'",
				"extends3": 1,
				"extends4": "Sqr",
				"extends5": "sqr(%)"
			}, {
				"$uuid": 13,
				"$serverIndex": 13,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FSUM",
				"fc2": "FSUM",
				"fc1": "FARITH",
				"fc3": "Somme  'sum(liste)'",
				"fc4": 1,
				"extends0": "FARITH",
				"extends1": "FSUM",
				"extends2": "Somme  'sum(liste)'",
				"extends3": 1,
				"extends4": "Sum",
				"extends5": "sum(%, )"
			}, {
				"$uuid": 14,
				"$serverIndex": 14,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FCHAINE",
				"fc2": "FCHAINE",
				"fc1": "",
				"fc3": "Chaînes",
				"fc4": 1,
				"extends0": "",
				"extends1": "FCHAINE",
				"extends2": "Chaînes",
				"extends3": 2,
				"extends4": "",
				"extends5": ""
			}, {
				"$uuid": 15,
				"$serverIndex": 15,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FLEFTD",
				"fc2": "FLEFTD",
				"fc1": "FCHAINE",
				"fc3": "Partie gauche  'left$(str,nbr)'",
				"fc4": 1,
				"extends0": "FCHAINE",
				"extends1": "FLEFTD",
				"extends2": "Partie gauche  'left$(str,nbr)'",
				"extends3": 2,
				"extends4": "Left$",
				"extends5": "left$(%, )"
			}, {
				"$uuid": 16,
				"$serverIndex": 16,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FLEN",
				"fc2": "FLEN",
				"fc1": "FCHAINE",
				"fc3": "Longueur  'len(str)'",
				"fc4": 1,
				"extends0": "FCHAINE",
				"extends1": "FLEN",
				"extends2": "Longueur  'len(str)'",
				"extends3": 2,
				"extends4": "Len",
				"extends5": "len(%)"
			}, {
				"$uuid": 17,
				"$serverIndex": 17,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FMIDD",
				"fc2": "FMIDD",
				"fc1": "FCHAINE",
				"fc3": "Sous-chaîne 'mid$(str,nbr,nbr)'",
				"fc4": 1,
				"extends0": "FCHAINE",
				"extends1": "FMIDD",
				"extends2": "Sous-chaîne 'mid$(str,nbr,nbr)'",
				"extends3": 2,
				"extends4": "Mid$",
				"extends5": "mid$(%, , )"
			}, {
				"$uuid": 18,
				"$serverIndex": 18,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FNUMD",
				"fc2": "FNUMD",
				"fc1": "FCHAINE",
				"fc3": "Transformation en chaîne 'num$(nbr)'",
				"fc4": 1,
				"extends0": "FCHAINE",
				"extends1": "FNUMD",
				"extends2": "Transformation en chaîne 'num$(nbr)'",
				"extends3": 2,
				"extends4": "Num$",
				"extends5": "num$(%)"
			}, {
				"$uuid": 19,
				"$serverIndex": 19,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FRIGHTD",
				"fc2": "FRIGHTD",
				"fc1": "FCHAINE",
				"fc3": "Partie droite 'right$(str,nbr)'",
				"fc4": 1,
				"extends0": "FCHAINE",
				"extends1": "FRIGHTD",
				"extends2": "Partie droite 'right$(str,nbr)'",
				"extends3": 2,
				"extends4": "Right$",
				"extends5": "right$(%, )"
			}, {
				"$uuid": 20,
				"$serverIndex": 20,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FTOLOWER",
				"fc2": "FTOLOWER",
				"fc1": "FCHAINE",
				"fc3": "Mise en minuscules  'tolower(str)'",
				"fc4": 1,
				"extends0": "FCHAINE",
				"extends1": "FTOLOWER",
				"extends2": "Mise en minuscules  'tolower(str)'",
				"extends3": 2,
				"extends4": "Tolower",
				"extends5": "tolower(%)"
			}, {
				"$uuid": 21,
				"$serverIndex": 21,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FTOUPPER",
				"fc2": "FTOUPPER",
				"fc1": "FCHAINE",
				"fc3": "Mise en majuscules 'toupper(str)'",
				"fc4": 1,
				"extends0": "FCHAINE",
				"extends1": "FTOUPPER",
				"extends2": "Mise en majuscules 'toupper(str)'",
				"extends3": 2,
				"extends4": "Toupper",
				"extends5": "toupper(%)"
			}, {
				"$uuid": 22,
				"$serverIndex": 22,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FVAL",
				"fc2": "FVAL",
				"fc1": "FCHAINE",
				"fc3": "Transformation en nombre  'val(str)'",
				"fc4": 1,
				"extends0": "FCHAINE",
				"extends1": "FVAL",
				"extends2": "Transformation en nombre  'val(str)'",
				"extends3": 2,
				"extends4": "Val",
				"extends5": "val(%)"
			}, {
				"$uuid": 23,
				"$serverIndex": 23,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FDATE",
				"fc2": "FDATE",
				"fc1": "",
				"fc3": "Dates",
				"fc4": 1,
				"extends0": "",
				"extends1": "FDATE",
				"extends2": "Dates",
				"extends3": 3,
				"extends4": "",
				"extends5": ""
			}, {
				"$uuid": 24,
				"$serverIndex": 24,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FADDMONTH",
				"fc2": "FADDMONTH",
				"fc1": "FDATE",
				"fc3": "Ajout d'un nombre de mois 'addmonth(date,nbr)'",
				"fc4": 1,
				"extends0": "FDATE",
				"extends1": "FADDMONTH",
				"extends2": "Ajout d'un nombre de mois 'addmonth(date,nbr)'",
				"extends3": 3,
				"extends4": "Addmonth",
				"extends5": "addmonth(%, )"
			}, {
				"$uuid": 25,
				"$serverIndex": 25,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FDAYD",
				"fc2": "FDAYD",
				"fc1": "FDATE",
				"fc3": "Renvoie le jour sous forme de chaine  'day$(str)'",
				"fc4": 1,
				"extends0": "FDATE",
				"extends1": "FDAYD",
				"extends2": "Renvoie le jour sous forme de chaine  'day$(str)'",
				"extends3": 3,
				"extends4": "Day$",
				"extends5": "day$(%)"
			}, {
				"$uuid": 26,
				"$serverIndex": 26,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FDAY",
				"fc2": "FDAY",
				"fc1": "FDATE",
				"fc3": "Renvoie le numéro du jour (1 à 31)  'day(nbr)'",
				"fc4": 1,
				"extends0": "FDATE",
				"extends1": "FDAY",
				"extends2": "Renvoie le numéro du jour (1 à 31)  'day(nbr)'",
				"extends3": 3,
				"extends4": "Day",
				"extends5": "day(%)"
			}, {
				"$uuid": 27,
				"$serverIndex": 27,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FDAYN",
				"fc2": "FDAYN",
				"fc1": "FDATE",
				"fc3": "Renvoie le numéro du jour (1 à 7)  'dayn(nbr)'",
				"fc4": 1,
				"extends0": "FDATE",
				"extends1": "FDAYN",
				"extends2": "Renvoie le numéro du jour (1 à 7)  'dayn(nbr)'",
				"extends3": 3,
				"extends4": "Dayn",
				"extends5": "dayn(%)"
			}, {
				"$uuid": 28,
				"$serverIndex": 28,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FEOMONTH",
				"fc2": "FEOMONTH",
				"fc1": "FDATE",
				"fc3": "Fin de mois  'eomonth(date)'",
				"fc4": 1,
				"extends0": "FDATE",
				"extends1": "FEOMONTH",
				"extends2": "Fin de mois  'eomonth(date)'",
				"extends3": 3,
				"extends4": "Eomonth",
				"extends5": "eomonth(%)"
			}, {
				"$uuid": 29,
				"$serverIndex": 29,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FGDATD",
				"fc2": "FGDATD",
				"fc1": "FDATE",
				"fc3": "Construction d'une date  'gdat$(nbr,nbr,nbr)'",
				"fc4": 1,
				"extends0": "FDATE",
				"extends1": "FGDATD",
				"extends2": "Construction d'une date  'gdat$(nbr,nbr,nbr)'",
				"extends3": 3,
				"extends4": "Gdat$",
				"extends5": "gdat$(%, , )"
			}, {
				"$uuid": 30,
				"$serverIndex": 30,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FMONTHD",
				"fc2": "FMONTHD",
				"fc1": "FDATE",
				"fc3": "Renvoie le mois en alpha  'month$(nbr)'",
				"fc4": 1,
				"extends0": "FDATE",
				"extends1": "FMONTHD",
				"extends2": "Renvoie le mois en alpha  'month$(nbr)'",
				"extends3": 3,
				"extends4": "Month$",
				"extends5": "month$(%)"
			}, {
				"$uuid": 31,
				"$serverIndex": 31,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FMONTH",
				"fc2": "FMONTH",
				"fc1": "FDATE",
				"fc3": "Renvoie le numéro de mois  'month(date)'",
				"fc4": 1,
				"extends0": "FDATE",
				"extends1": "FMONTH",
				"extends2": "Renvoie le numéro de mois  'month(date)'",
				"extends3": 3,
				"extends4": "Month",
				"extends5": "month(%)"
			}, {
				"$uuid": 32,
				"$serverIndex": 32,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FWEEK",
				"fc2": "FWEEK",
				"fc1": "FDATE",
				"fc3": "Renvoie le numéro de semaine 'week(date)'",
				"fc4": 1,
				"extends0": "FDATE",
				"extends1": "FWEEK",
				"extends2": "Renvoie le numéro de semaine 'week(date)'",
				"extends3": 3,
				"extends4": "Week",
				"extends5": "week(%)"
			}, {
				"$uuid": 33,
				"$serverIndex": 33,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FYEAR",
				"fc2": "FYEAR",
				"fc1": "FDATE",
				"fc3": "Renvoie l'année 'year(date)'",
				"fc4": 1,
				"extends0": "FDATE",
				"extends1": "FYEAR",
				"extends2": "Renvoie l'année 'year(date)'",
				"extends3": 3,
				"extends4": "Year",
				"extends5": "year(%)"
			}, {
				"$uuid": 34,
				"$serverIndex": 34,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FDIVERS",
				"fc2": "FDIVERS",
				"fc1": "",
				"fc3": "Divers",
				"fc4": 1,
				"extends0": "",
				"extends1": "FDIVERS",
				"extends2": "Divers",
				"extends3": 4,
				"extends4": "",
				"extends5": ""
			}, {
				"$uuid": 35,
				"$serverIndex": 35,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FFIND",
				"fc2": "FFIND",
				"fc1": "FDIVERS",
				"fc3": "Recherche d'une valeur  'find(str,str)'",
				"fc4": 1,
				"extends0": "FDIVERS",
				"extends1": "FFIND",
				"extends2": "Recherche d'une valeur  'find(str,str)'",
				"extends3": 4,
				"extends4": "Find",
				"extends5": "find(%, )"
			}, {
				"$uuid": 36,
				"$serverIndex": 36,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FFORMATD",
				"fc2": "FFORMATD",
				"fc1": "FDIVERS",
				"fc3": "Formatage  'format$(str,str)'",
				"fc4": 1,
				"extends0": "FDIVERS",
				"extends1": "FFORMATD",
				"extends2": "Formatage  'format$(str,str)'",
				"extends3": 4,
				"extends4": "Format$",
				"extends5": "format$(%, )"
			}, {
				"$uuid": 37,
				"$serverIndex": 37,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FINSTR",
				"fc2": "FINSTR",
				"fc1": "FDIVERS",
				"fc3": "Recherche d'une sous-chaîne  'instr(str,nbr,str)'",
				"fc4": 1,
				"extends0": "FDIVERS",
				"extends1": "FINSTR",
				"extends2": "Recherche d'une sous-chaîne  'instr(str,nbr,str)'",
				"extends3": 4,
				"extends4": "Instr",
				"extends5": "instr(%, , )"
			}, {
				"$uuid": 38,
				"$serverIndex": 38,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FPAT",
				"fc2": "FPAT",
				"fc1": "FDIVERS",
				"fc3": "Conformité à un modèle  'pat(str,str)'",
				"fc4": 1,
				"extends0": "FDIVERS",
				"extends1": "FPAT",
				"extends2": "Conformité à un modèle  'pat(str,str)'",
				"extends3": 4,
				"extends4": "Pat",
				"extends5": "pat(%, )"
			}, {
				"$uuid": 39,
				"$serverIndex": 39,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FSPACED",
				"fc2": "FSPACED",
				"fc1": "FDIVERS",
				"fc3": "Chaîne d'espaces  'space$(nbr)'",
				"fc4": 1,
				"extends0": "FDIVERS",
				"extends1": "FSPACED",
				"extends2": "Chaîne d'espaces  'space$(nbr)'",
				"extends3": 4,
				"extends4": "Space$",
				"extends5": "space$(%)"
			}, {
				"$uuid": 40,
				"$serverIndex": 40,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FSTRINGD",
				"fc2": "FSTRINGD",
				"fc1": "FDIVERS",
				"fc3": "Répétition chaîne  'string$(nbr,str)'",
				"fc4": 1,
				"extends0": "FDIVERS",
				"extends1": "FSTRINGD",
				"extends2": "Répétition chaîne  'string$(nbr,str)'",
				"extends3": 4,
				"extends4": "String$",
				"extends5": "string$(%, )"
			}, {
				"$uuid": 41,
				"$serverIndex": 41,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FVIREBLC",
				"fc2": "FVIREBLC",
				"fc1": "FDIVERS",
				"fc3": "Suppression des blancs superflus  'vireblc(str,nbr)'",
				"fc4": 1,
				"extends0": "FDIVERS",
				"extends1": "FVIREBLC",
				"extends2": "Suppression des blancs superflus  'vireblc(str,nbr)'",
				"extends3": 4,
				"extends4": "Vireblc",
				"extends5": "vireblc(%, )"
			}, {
				"$uuid": 42,
				"$serverIndex": 42,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FONCTION",
				"fc2": "FONCTION",
				"fc1": "",
				"fc3": "Fonctions",
				"fc4": 1,
				"extends0": "",
				"extends1": "FONCTION",
				"extends2": "Fonctions",
				"extends3": 275,
				"extends4": "Aide sur fonction",
				"extends5": "Fonction"
			}, {
				"$uuid": 43,
				"$serverIndex": 43,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FONCTION.AFNC/ACTIV",
				"fc2": "FONCTION.AFNC/ACTIV",
				"fc1": "FONCTION",
				"fc3": "Valeur d'un code activité",
				"fc4": 1,
				"extends0": "FONCTION",
				"extends1": "FONCTION.AFNC/ACTIV",
				"extends2": "Valeur d'un code activité",
				"extends3": 275,
				"extends4": "AFNC_ACTIV",
				"extends5": "func AFNC.ACTIV(COD)"
			}, {
				"$uuid": 44,
				"$serverIndex": 44,
				"$rawV": {
					"fc1": {},
					"fc2": {},
					"fc3": {}
				},
				"$key": "122~FONCTION.AFNC/CHEF",
				"fc2": "FONCTION.AFNC/CHEF",
				"fc1": "FONCTION",
				"fc3": "Supérieur hiérarchique",
				"fc4": 1,
				"extends0": "FONCTION",
				"extends1": "FONCTION.AFNC/CHEF",
				"extends2": "Supérieur hiérarchique",
				"extends3": 275,
				"extends4": "AFNC_CHEF",
				"extends5": "func AFNC.CHEF(USER,N)"
			}],
			"$isPartialDelta": false
		});
	},
	applyFieldMetaData: function($bind, value, $field) {

	}
});
