// syracuse-ui/lib/tests/misc/searchVignetteTest.js
"use strict";
var helpers = require("syracuse-core/lib/helpers");
var TestPage = require("syracuse-ui/lib/tests/testPage").TestPage;

function SearchVignetteTest(){
}

exports.SearchVignetteTest = helpers.defineClass(SearchVignetteTest, TestPage, {
    _initiliazeTestPage: function(){
        var self = this;
        self.$prototype.$links.$select = {
            $title: "Add vignette",
			$url: "/sdata/syracuse/search/syracuse/$search?representation=queryResult.$search&q=*&source=data&filters=~(type~(~'portlet))"
        };
        
        self.$item.$layout.$items = [{
            $category: "section",
            $title: "Vignette adding test",
            $layout: {
                $items: [{
                    $category: "link",
                    $bind: "$select"
                }]
            }
        }];
        
    },
    onMenuClick: function(menu){
		document.site.showMessage({
			$message: "menu clicked",
			$type: "info",
			$title: "vignette adding test",
			$buttons: "ok"
		});
        return false;
    }
});
