// syracuse-ui/lib/tests/misc/searchVignetteTest.js
"use strict";
var helpers = require("syracuse-core/lib/helpers");
var TestPage = require("syracuse-ui/lib/tests/testPage").TestPage;
var SearchVignettePage = require("syracuse-ui/lib/site/search/vignette/searchVignettePage").SearchVignettePage;

function SearchVignetteTest(){
}

exports.SearchVignetteTest = helpers.defineClass(SearchVignetteTest, TestPage, {
    _initiliazeTestPage: function(){
        var self = this;
        self.$prototype.$links.$select = {
            $title: "Select vignette",
            $url: "/sdata/syracuse/collaboration/syracuse/landingVignetteSelects/$template/$workingCopies?representation=landingVignetteSelect.$edit",
			$method: "POST"
        };
        
        self.$item.$layout.$items = [{
            $category: "section",
            $title: "Vignette adding test",
            $layout: {
                $items: [{
                    $category: "link",
                    $bind: "$select"
                }]
            }
        }];
        
    },
    onMenuClick: function(menu){
        (new SearchVignettePage())._openPage(menu);
        //(new SearchVignettePage()).loadData(menu);
        /*document.site.showMessage({
         $message: "menu clicked",
         $type: "info",
         $title: "vignette adding test",
         $buttons: "ok"
         });*/
        return false;
    }
});
