// syracuse-ui/lib/filter/criteria.js
"use strict";
var _parser = require('syracuse-sdata-parser');

function _formatFusionCriteria(options) {
	var values = ["", options.value.replace(/\**$/, ""), ""];
	switch (options.operator) {
		case "none":
			break;
		case "gt":
			values[0] = ">";
			break;
		case "ge":
			values[0] = ">=";
			break;
		case "lt":
			values[0] = "<";
			break;
		case "le":
			values[0] = "<=";
			break;
		case "eq":
			values[0] = "=";
			break;
		case "ne":
			values[0] = "<>";
			break;
		case "like_s":
			values[2] = "*";
			break;
		case "like":
			values[0] = "*";
			values[2] = "*";
			break;
		case "between":
			break;
	}
	return values.join("");
}


function _formatFieldCriteria(options) {
	switch (options.$type) {
		case "application/x-reference":
			return _formatFieldCriteria({
				field: options.field,
				$bind: options.$bind,
				$type: options.$filterField.$type,
				value: options.field.getInputValue(),
				getRawValue: options.getRawValue
			});
		case "application/x-choice":
			var $field = options.field.$field;
			return _formatFieldCriteria({
				field: options.field,
				$bind: options.$bind,
				$type: ($field && $field.$value && $field.$value.$type) || "application/x-string",
				value: options.field.articleParent.dataset[options.$bind],
				getRawValue: options.getRawValue
			});
		case "application/x-boolean":
		case "application/x-integer":
		case "application/x-decimal":
		case "application/x-quantity":
			return options.field.articleParent.dataset[options.$bind];
		case "application/x-date":
			options.value = options.field.articleParent.dataset[options.$bind];
			return options.getRawValue ? options.value : ("@" + options.value + "@");
	}
	return options.getRawValue ? options.value : ("'" + (options.value || "") + "'");
}

exports.format = function(options) {
	options.value = options.field.getInputValue();
	if (options.field.page.isFusionPage) {
		return _formatFusionCriteria(options);
	}
	if (options.value != null) {
		switch (options.operator) {
			case "like":
				options.getRawValue = true;
				options.value = "'%" + ("" + _formatFieldCriteria(options)).replace(/'/g, "''") + "%'";
				break;
			case "like_s":
				options.getRawValue = true;
				options.value = "'" + ("" + _formatFieldCriteria(options)).replace(/'/g, "''") + "%'";
				options.operator = "like";
				break;
			case "between":
				var highValue = options.field.articleParent.dataset[options.$bind + "$High"];
				if (value && !highValue) {
					options.value = _formatFieldCriteria(options);
					options.operator = "ge";
				} else {
					if (!value && highValue) {
						options.value = highValue;
						options.value = _formatFieldCriteria(options);
						operator = "le";
					} else {
						if (value && highValue) {
							var minValue = _formatFieldCriteria(options);
							options.value = highValue;
							options.value = minValue + " and " + _formatFieldCriteria(options);
						} else {
							return "";
						}
					}
				}
				break;
			case "empty":
				options.value = "''";
				options.operator = "eq";
				break;
			case "notempty":
				options.value = "''";
				options.operator = "ne";
				break;
			default:
				options.value = _formatFieldCriteria(options);
				break;
		}

		var $field = options.field.$field;
		// handling specific reference field case (fix SAM 97863)
		if ($field) {
			switch ($field.$type) {
				case "application/x-reference":
					if (options.value == "''" && options.$filterCode.indexOf(".") != -1) {
						options.value = null;
						options.$filterCode = options.field.$item.$bind;
					}
					break;
			}
			var list = options.field.articleParent.list;
			if (list && !list.isClientFetch && $field.$capabilities.indexOf("filter_upper") >= 0) {
				options.value = options.value.toUpperCase();
				options.$filterCode = "upper(" + options.$filterCode + ")";
			}
		}
		return "(" + options.$filterCode + " " + options.operator + " " + options.value + ")";
	}
};

function _getCriteria(exp, keys, parentExp) {
	var criteria;
	for (var ii = 0, jj = exp.children.length; !criteria && ii < jj; ii++) {
		var child = exp.children[ii];
		if (child.type == "identifier" && keys.indexOf(child.value) >= 0) {
			var root = jj > 1 ? exp : (parentExp && parentExp.children.length > 1) ? parentExp : null;
			if (root) {
				if (keys.length == 1 && (root.children[1].type != "identifier")) {
					criteria = {
						keys: keys,
						$value: root.children[1].value,
						$operator: root.value.text
					};
					switch (criteria.$operator) {
						case "like":
						case "like_s":
							if (criteria.$value.length) {
								var first = criteria.$value[0],
									last = criteria.$value[criteria.$value.length - 1];
								if (criteria.$operator == "like") {
									if (first != "%") {
										if (last == "%") {
											criteria.$operator = "like_s";
										}
									} else {
										criteria.$value = criteria.$value.slice(1);
									}
								}
								if (last == "%") {
									criteria.$value = criteria.$value.slice(0, -1);
								}
							}
							break;
						case "eq":
						case "ne":
							if (!criteria.$value.length) {
								criteria.$operator = (criteria.$operator == "eq") ? "empty" : "notempty";
							}
							break;

					}
				} else {
					if (root.children.length == keys.length && root.value.text == ".") {
						for (var mm = 0, kk = keys.length; mm < kk; mm++) {
							if (root.children[mm] != keys[mm]) {
								break;
							}
						}
						if (mm == kk) {
							criteria = {
								keys: keys,
								$value: parentExp.children[1].value,
								$operator: parentExp.value.text
							};
						}
					}
				}
			}
		} else {
			criteria = child.children && _getCriteria(child, keys, exp);
		}
	}
	return criteria;
}

exports.parse = function(where, keys) {
	var exp = _parser.parse(where);
	return (exp && keys) ? _getCriteria(exp, keys) : null;
};
