// syracuse-ui/lib/filter/operators.js
"use strict";

function _clickOperator(event) {
	var field = this.parent;
	exports.setFieldOperator(field, undefined, this.operator);
	field.page.filter_popup && field.page.filter_popup.close();
	field.page.externalAdapter.onFilterEvent({
		operator: field.filter_selectedOperator,
		list: field.articleParent.list,
		filter: field,
		event: event,
		type: "operatorchange",
		doEvent: function() {
			var record = field.articleParent;
			var value = field.getInputValue();
			if (field.filter_selectedOperator != "none") {
				if (value != null && value != "") {
					if (field.filter_selectedOperator == "empty" || field.filter_selectedOperator == "notempty") {
						syra_filter.setValue(field, null, field.filter_selectedOperator);
					}
					record.triggerFilter();
				} else {
					if (field.filter_selectedOperator == "empty" || field.filter_selectedOperator == "notempty") {
						record.triggerFilter();
					}
				}
			} else {
				if (value != null && value != "") {
					syra_filter.setValue(field, null, field.filter_selectedOperator);
					syra_form.update(field, null);
				} else {
					record.triggerFilter();
				}
			}
		}
	});
}

function _clickShowList(event) {
	var field = this.parent;
	var record = field.articleParent;
	record.page.externalAdapter.onFilterEvent({
		operator: field.filter_selectedOperator,
		list: record.list,
		event: event,
		type: "operatorclick",
		doEvent: function() {
			if (field.page.filter_popup) {
				field.page.filter_popup.close();
			} else {
				field.ensureFilterSetting && field.ensureFilterSetting();
				var ul = syra_dom.ul("s-over-popup s-filter-operators");
				ul.syraItem = field.id;
				exports.assignToField(field);
				for (var ii = 0, jj = field.filter_operators.length; ii < jj; ii++) {
					var op = field.filter_operators[ii];
					var btn = syra_button.add({
						parent: field,
						slot: syra_dom.li("s-filter-operator", ul),
						text: syra_local["flFilter_" + op],
						css: "s-mn-link",
						fontIcon: "filter_" + op,
						click: _clickOperator,
						operator: op
					});
					if (field.filter_selectedOperator == op) {
						btn.link.className += " s-selected";
					}
				}
				field.page.filter_popup = syra_over.openPopup(record.boxParent, {
					content: record,
					slot: ul,
					picker: field.layoutSlot,
					position: {
						my: "left top",
						at: "left bottom",
						of: field.layoutSlot
					},
					close: function() {
						if (field.page) {
							field.page.filter_popup = null;
						}
					}
				});
			}
		}
	});
}

function _getFieldOperators(field) {
	if (field.page.isFusionPage) {
		switch ((field.$filterField && field.$filterField.$type) || field.$field.$type) {
			case "application/x-date":
			case "application/x-time":
			case "application/x-datetime":
				return ["none", "ge", "gt", "lt", "le", "eq", "ne"];
			case "application/x-string":
				return ["none", "like_s", "like", "gt", "ge", "lt", "le", "eq", "ne"];
			case "application/x-choice":
			case "application/x-real":
			case "application/x-decimal":
			case "application/x-integer":
			case "application/x-quantity":
				return ["none", "eq", "gt", "ge", "lt", "le", "ne"];
			default:
				return ["none", "eq", "ne"];
		}
	} else {
		switch ((field.$filterField && field.$filterField.$type) || field.$field.$type) {
			case "application/x-date":
			case "application/x-time":
			case "application/x-datetime":
				var operators = ["none", "ge", "gt", "lt", "le", "eq", "ne", "empty", "notempty"];
				if (field.$field.$type != "application/x-reference") {
					operators.push("between");
				}
				return operators;
			case "application/x-string":
				return ["none", "like_s", "like", "gt", "ge", "lt", "le", "eq", "ne", "empty", "notempty"];
			case "application/x-binary":
			case "application/x-document":
			case "image":
				return ["none", "empty", "notempty"];
			case "application/x-boolean":
				return ["none", "eq", "ne"];
			case "application/x-decimal":
			case "application/x-integer":
			case "application/x-quantity":
				if (field.$field.$type != "application/x-reference") {
					return ["none", "gt", "ge", "lt", "le", "eq", "ne", "between"];
				}
				return ["none", "eq", "ne"];
		}
		return ["none", "eq", "ne", "empty", "notempty"];
	}
}

exports.assignToField = function(field) {
	return field.filter_operators || (field.filter_operators = _getFieldOperators(field));
};

exports.getDefault = function(field) {
	return exports.assignToField(field)[1];
};

exports.addPicker = function(field, css) {
	field.filter_picker = syra_button.add({
		parent: field,
		text: syra_local.flfilter_choiceOperator,
		css: css,
		iconOnly: true,
		fontIcon: "filter_none",
		click: _clickShowList
	});
	field.filter_picker.link.removeAttribute("href");
	field._dataValue.insertBefore(field.filter_picker.link, field._dataValue.firstChild);
};

exports.setFieldOperator = function(field, value, operator) {
	if (operator === undefined) {
		operator = field.filter_selectedOperator;
		if (value == null || value == "") {
			operator = "none";
		} else {
			if (field.$field && operator == "none") {
				operator = exports.getDefault(field);
			}
		}
	}
	syra_dom.toggleClass(field.filter_slot, field.articleParent.filter_criteria_css + "-active", operator != "none");
	syra_button.setText(field.filter_picker, "", "filter_" + (field.filter_selectedOperator = operator));
};
