// syracuse-ui/lib/page/landing/landingPage.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var DesktopPage = require('syracuse-ui/lib/page/desktopPage').DesktopPage;
var LandingPageBar = require('./landingPageBar').LandingPageBar;
var LandingPageList = require('./landingPageList').LandingPageList;

function LandingPage(){
}

exports.LandingPage = helpers.defineClass(LandingPage, DesktopPage, {
    designItem: function(){
    },
    ensureDesignedItemVisibility: function(item, event){
        /* if (item.menusSlot) {
         item.menusSlot.style.visibility = "visible";
         }
         if (item.reorderPicker) {
         item.reorderPicker.style.visibility = "visible";
         }
         return;
         */
        this.showOnEnter(item.menusSlot, event);
        this.showOnEnter(item.reorderPicker, event);
    },
    onBarEvent: function(picker, event){
        if (picker.className.indexOf("s-nav-modulebar") >= 0) {
            this.bar.onClickPicker(picker, event);
        }
        else {
            DesktopPage.prototype.onBarEvent.call(this, picker, event);
        }
    },
    showOnEnter: function(domItem, event){
        if (domItem) {
            domItem.style.visibility = (event && event.type == "mouseenter") ? "visible" : "hidden";
        }
    },
    getDataUrl: function(article){
        article = article || this;
        var $parsedUrl = article.parseExpression(article.dataset.$url || article.$prototype.$url);
        return $parsedUrl || this.$prototype.$representationUrl;
    },
    notifyChangeToServer: function(sendBag, articleSender){
        var self = this;
        articleSender = articleSender || this;
        if (!self.disposed && !self.isServerNotifyDisabled) {
            if (self.$prototype.$representationUrl) {
                sendBag.$url = self.getDataUrl(articleSender);
                sendBag.$etag = articleSender.dataset.$etag;
                document.controller.sendRequest(self, {
                    $location: {
                        $url: self.getDataUrl(articleSender)
                    },
                    data: sendBag,
                    method: "PUT",
                    $etag: sendBag.$etag
                }, function(data, response, requestUrl){
                    if (!articleSender.disposed) {
                        articleSender.applyChange(data, response, requestUrl);
                        self.ensurePageVisibility();
                    }
                });
            }
        }
    },
    _ensureMenus: function(){
        this.page.loadNewItem(this.headerCoreRight, {
            $noText: true,
            $category: "link",
            $skin: "s-navpage-designer-link",
            $bind: this.isAdminMode ? "$details" : "$edit",
            $icon: {
                $mode: "icon",
                $path: "page/s-navpage-designer-"
            }
        }, this);
        
    },
    ensureDefaultArticle: function($article, $prototype){
        return {
            $layout: {
                $items: []
            }
        };
    },
    createField: function($field, $item, boxParent, $class){
        switch ($item.$bind) {
            case "rolePages":
            case "userPages":
                return new LandingPageList();
        }
        return null;
    },
    ensureDesignerOpenerVisibility: function(){
        document.site.enablePageDesign(false);
    },
    loadBox: function(initData, $initDiagnoses){
        this.isAdminMode = this.$prototype.$editMode == "admin";
		this.modules = {};
        this.$isEditMode = true;
        this.isLandingPage = true;
        this.$skin = this.$item.$skin || "s-navpage";
        this.$defaultSkinSection = this.$defaultSkinBlock ="s-navpage-h1";
        this._defaultTitle = this.localize.navigPageTitle;
        this.$autoFetch = false;
        var data = {};
        if (initData) {
            var $keys = Object.keys(initData);
            for (var ii = 0, jj = $keys.length; ii < jj; ii++) {
                if ($keys[ii] != "$prototype" && $keys[ii] != "$article") {
                    data[$keys[ii]] = initData[$keys[ii]];
                }
            }
        }
        DesktopPage.prototype.loadBox.call(this, data, $initDiagnoses);
        if (!this.bar) {
            (this.bar = new LandingPageBar()).load(this);
        }
        this.ensurePageVisibility();
    },
    getModuleList: function(){
        return this.boundFields && this.boundFields.modules && this.boundFields.modules[0];
    },
    changeMode: function($url){
        var self = this;
        var $newEditmode = this.isAdminMode ? "user" : "admin";
        document.controller.parseUrl($url).sendRequest(null, function($representation, response, $url){
            $representation.$pageRepresentationURL = $url;
            $representation.$prototype.$representationUrl = self.$prototype.$representationUrl;
            $representation.$prototype.$editMode = $newEditmode;
            document.site.onMainPageChange({
                $representation: $representation,
                $urlParts: self.$urlParts,
                httpQuery: self.openerHttpQuery,
                layoutSlot: self.layoutSlot
            });
            
        }, function(){
        
        });
    },
    onMenuClick: function(menuItem){
        var self = this;
        if (this.isAdminMode && menuItem.isNavigationMenuItem) {
            return false;
        }
        if (!menuItem.$isAction) {
            if (menuItem.articleParent == this) {
                if (menuItem.$item.$bind == "$edit" ||
                menuItem.$item.$bind == "$details") {
                    this.changeMode(menuItem.$url);
                    return false;
                }
            }
            switch (menuItem.$item.$bind) {
                case "$create":
                case "$edit":
                    self.openDialog({
                        article: menuItem.articleParent,
                        $url: menuItem.$url,
                        $method: menuItem.$method,
                        onSave: function(menuItem, dialog){
                            dialog.options.article.notifyClientSave(dialog._content.dataset);
                            dialog.close(true);
                            return false;
                        },
                        onValidate: function(page){
                            if (page.validateFields()) {
                                page.clickMenu("$save");
                            }
                            return false;
                        },
                        onClose: function(isCanceled, dispose){
                            return true;
                        }
                    });
                    return false;
            }
        }
        return true;
    },
    
    notifySubModuleDataChange: function(field, value, notifyServerChange){
        var subModule = field.articleParent;
        while (subModule && subModule != this) {
            if (subModule.isSubModuleRecord) {
                var sendBag = this.ensureSendBag(subModule, subModule);
                sendBag.saveDataChange(subModule, field.getArticle(), field.$item.$bind, field.parentVariantField ? field.parentVariantField.saveVariantValue(value) : value);
                if (notifyServerChange !== false) {
                    this.notifyChangeToServer(sendBag, subModule);
                }
                return true;
            }
            subModule = subModule.articleParent;
        }
        return false;
    },
    notifyDataChange: function(field, value, notifyServerChange){
        if (field.isSubModuleList || field.articleParent.isSubModuleRecord) {
            var sendBag = this.ensureSendBag(field.articleParent, field.articleParent);
            sendBag.saveDataChange(field.articleParent, field.getArticle(), field.$item.$bind, field.parentVariantField ? field.parentVariantField.saveVariantValue(value) : value);
            if (notifyServerChange !== false) {
                this.notifyChangeToServer(sendBag, field.articleParent);
            }
        }
        else {
            if (!this.notifySubModuleDataChange(field, value, notifyServerChange)) {
                DesktopPage.prototype.notifyDataChange.call(this, field, value);
            }
        }
    },
    getNavigationPath: function(navigationMenuItem){
        var path = {
            menuItem: navigationMenuItem.currentValue.$uuid
        };
        var articleParent = navigationMenuItem.articleParent;
        while (articleParent) {
            articleParent = articleParent.articleParent;
        }
        return path;
    },
    onItemEnterLeave: function(event){
        if (this.bar) {
            if (event.currentTarget == this.bar.barBody) {
                var modules = this.getModuleList();
                if (modules) {
                    if (!this.isAdminMode) {
                        if (modules.topbar) {
                            modules.topbar.style.visibility = "hidden";
                        }
                        if (modules.bottombar) {
                            modules.bottombar.style.visibility = "hidden";
                        }
                    }
                    else {
                        if (event) {
                            var visible = event.type == "mouseenter" ? "visible" : "hidden";
                            if (modules.topbar) {
                                modules.topbar.style.visibility = visible;
                            }
                            if (modules.bottombar) {
                                modules.bottombar.style.visibility = visible;
                            }
                        }
                    }
                    this.page.ensureDesignedItemVisibility(modules, event);
                }
            }
        }
    },
    dispose: function(){
        if (this.bar) {
            this.bar.dispose();
        }
        this.modules = this.bar = null;
        DesktopPage.prototype.dispose.call(this);
    }
    
});
