// syracuse-ui/lib/page/landing/landingPage.js
"use strict";
var helpers = require('syracuse-core').helpers;
var DesktopPage = require('syracuse-ui/lib/page/desktopPage').DesktopPage;
var SideBar = require("syracuse-ui/lib/page/aside/sideBar").SideBar;
var Designer = require("syracuse-ui/lib/authoring/landingPageDesigner").Designer;

var _newldp = require("./newLandingPage");

function LandingPage() {}

exports.LandingPage = helpers.defineClass(LandingPage, DesktopPage, {
	onExecuteMenuResponse: function($bind, record, responseData) {
		if ($bind == "$delete") {
			var self = this;
			var list = record.list;
			var delta = {};
			var index = syra_dom.getNodeIndex(record.domItem);
			delta[record.list.$item.$bind] = [{
				$uuid: record.$uuid,
				$isDeleted: true,
				$index: record.$serverIndex
			}];
			self.page.applyChange(delta);
			setTimeout(function() {
				if (!self.ldpRecord) {
					index = Math.min(index, list._core.childNodes.length - 1);
					if (index >= 0) {
						record = syra_item.findItem(list._core.childNodes[index]);
						if (record) {
							list.selector.selectRecord(record, true);
							return;
						}
					}
					self._ensureSelection();
				}
			}, 10);
		}
	},
	saveLayoutContent: function() {
		var record = this.ldpRecord;
		this.addTabBtn && this.layoutContent.ensureTabsSlot().appendChild(this.addTabBtn.link);
		syra_locker.lock();
		var sending = {
			$etag: 1,
			$url: syra_url.formatMenuUrl(record, record.ldp_getSaveLink())
		};
		sending.content = syra_layout.cleanBeforeSave(this.$item);
		delete sending.content.$menus;
		delete sending.content.$isModel;
		syra_ajax.put({
			page: this,
			url: sending.$url,
			data: sending,
			success: function(data, response, requestUrl) {
				try {
					if (!record.disposed && data.content) {
						(record.dataset.stdLayout = record.dataset.stdLayout || {}).content = data.content;
					}
				} finally {
					syra_locker.unlock();
				}
			},
			error: function() {
				syra_locker.unlock();
				return true;
			}
		});
	},
	updateWorkingCopy: function(sendBag, article, onUpdateContent) {
		var self = this;
		article = article || this;
		if (!self.disposed && !self.isServerNotifyDisabled) {
			if (self.$prototype.$representationUrl) {
				sendBag.$url = self.getDataUrl(article);
				sendBag.$etag = article.dataset.$etag;
				syra_locker.lock();
				syra_ajax.put({
					page: self,
					url: self.getDataUrl(article),
					data: sendBag,
					$etag: sendBag.$etag,
					success: function(newData, response, requestUrl) {
						try {
							if (!article.disposed) {
								if (article.page != article) {
									syra_dataset.applyDelta(article.page, article.dataset, newData);
								}
								article.applyChange(newData, response, requestUrl);
								if (onUpdateContent) {
									delete self.ldpRecord;
									self.userPages.selector.selectRecord(article, true, true);
								} else {
									syra_layout.ensureArticleVisibility(self);
								}
							}
						} finally {
							syra_locker.unlock();
						}
					},
					error: function() {
						syra_locker.unlock();
						return true;
					}
				});
			}
		}
	},
	deleteOpenedTab: function() {
		var self = this;
		var box = self.layoutContent.getOpenedTab();
		if (box && box.layoutContent) {
			var vignettes = box.layoutContent.getFields();
			syra_alert.ask({
				$title: syra_local.ldpDeleteVignette,
				$message: syra_local.ldpConfirmTabDeleteVignette.replace("{0}", box.getTitle()),
				yes: function(response) {
					var uiids = [];
					for (var ii = 0, jj = vignettes.length; ii < jj; ii++) {
						uiids.push(vignettes[ii].$field.$vignetteId);
					}
					var fields = box.layoutContent.getFields();
					for (var ii = 0, jj = fields.length; ii < jj; ii++) {
						syra_item.remove(fields[ii], true, true);
					}
					box.layoutParent.removeItem(box, true);
					self.ldpRecord.ldp_deleteVignettes(uiids);
				}
			});
		}
	},
	getDataUrl: function(article) {
		article = article || this;
		var $parsedUrl = syra_expression.parse(article, article.dataset.$url || article.$prototype.$url);
		return $parsedUrl || this.$prototype.$representationUrl;
	},
	initializePage: function($itemPage) {
		return {
			$layout: {
				$items: []
			}
		};
	},
	saveLandingPreferences: function() {
		syra_preference.profile.save("landingpage", this.$profilePreference);
	},

	load: function(initData) {
		this.isMenuBarDisabled = true;
		this.isAutoInsertFieldDisabled = true;
		this.isBackButtonDisabled = true;
		this.mainPageDesignerAccess = "unknow";
		this.$profilePreference = syra_preference.profile.get("landingpage") || {};
		this.$profilePreference.tabs = this.$profilePreference.tabs || {};
		this.$defaultSkinBlock = "s-ldp-h2";
		this.$defaultSkinSection = "s-ldp-h1";
		this.$skin = "s-ldp";
		this.isLandingPage = true;
		syra_site.landingPage = this;
		this.$isEditMode = true;
		var data = {};
		if (initData) {
			var $keys = Object.keys(initData);
			for (var ii = 0, jj = $keys.length; ii < jj; ii++) {
				if ($keys[ii] != "$prototype" && $keys[ii] != "$article") {
					data[$keys[ii]] = initData[$keys[ii]];
				}
			}
		}
		DesktopPage.prototype.load.call(this, data);
		this.menuBar = new SideBar();
		this.menuBar.preferenceKey = "ldpbar";
		this.menuBar.options = {
			resizeDirection: "left",
			$skin: "s-ldpm-bar"
		};
		this.menuBar.slot = syra_dom.div("s-ldpm-bar-slot");
		this.menuBar.body = syra_dom.div("s-ldpm-bar-body");
		if (this.$facet != "$landing_edit") {
			this.domItem.insertBefore(this.menuBar.slot, this.domItem.firstChild);
		}
		this.menuBar.load(this);

		this.newRecordsMap = {};
		this.rolePageList = new _newldp.RolePageList(this);
		this.rolePageList.addChildren();


		if (syra_site.userProfile.hasDesignRight()) {
			this.userPageList = new _newldp.UserPageList(this);
			this.userPageList.addChildren();

		}
		var slot = syra_dom.div("s-ldp-head-btns");
		this.addVignetteBtn = syra_button.add({
			parent: this,
			slot: slot,
			isHidden: true,
			text: syra_local.ldpAddVignette,
			css: "s-ldp-btn",
			iconOnly: true,
			fontIcon: "create",
			shortCutTip: syra_shortCuts.tip.create,
			click: function() {
				this.parent._addVignette();
			}
		});
		this.refreshBtn = syra_button.add({
			parent: this,
			slot: slot,
			text: syra_local.ldpRefresh,
			css: "s-ldp-btn",
			iconOnly: true,
			fontIcon: "refresh",
			shortCutTip: syra_shortCuts.tip.refresh,
			click: function() {
				var items = this.parent.layoutContent.getFields();
				for (var ii = 0, jj = items.length; ii < jj; ii++) {
					var item = items[ii];
					item.refresh && item.refresh(true);
				}
			}
		});
		this.designBtn = syra_button.add({
			parent: this,
			slot: slot,
			text: syra_local.ldpOpenDesigner,
			css: "s-ldp-btn",
			iconOnly: true,
			fontIcon: "design",
			isHidden: true,
			click: function() {
				this.parent.designer.toggleDesigner();
			}
		});
		this.addTabBtn = syra_button.add({
			parent: this,
			isHidden: true,
			text: "+",
			title: syra_local.ldpAddTab,
			css: "s-ldp-add-tab",
			shortCutTip: syra_shortCuts.tip.newTab,
			click: function() {
				this.parent.createNewTab();
				this.parent.addVignetteBtn.click();
				this.parent.designer && this.parent.designer.endLayoutDisplay();
			}
		});
		this.header.appendChild(slot);
		this._ensureSelection();
	},
	_ensureSelection: function() {
		var record = this.newRecordsMap[this.$profilePreference.$selected || this.dataset.$selected];
		if (!record) {
			record = this.rolePageList.children.length && this.rolePageList.children[0];
			if (!record && this.userPageList) {
				record = this.userPageList.children.length && this.userPageList.children[0];
			}
		}
		record && record.activate(true);
	},
	onMenuClick: function(options) {
		var self = this;
		if (!options.menu.$isAction) {
			switch (options.menu.$sourceBind) {
				case "delete":
					return false;
				case "$create":
				case "$edit":
					syra_over.openModal(self, {
						article: options.menu.articleParent,
						$url: options.menu.$url,
						$method: options.menu.$method,
						onServerSaved: function(saveItem, modal) {
							syra_article.notifyClientSave(modal.options.article, modal.page.dataset);
							var $uuid = modal.page.dataset.$uuid;
							modal.close(true);
							var record = self.userPages.recordsMap[$uuid];
							if (options.menu.$sourceBind == "$create") {
								self.userPages.selector.selectRecord(record, true);
								self.userPages.builder.saveReorder();
								self.addTabBtn.click();
							} else {
								self.userPages.selector.selectRecord(record, false);
								self.userPages.selector.selectRecord(record, true);
							}
							return false;
						},
						close: function(isCanceled, dispose) {
							return true;
						}
					});
					return false;
			}
		}
		return true;
	},


	section_onAfterOpentTab: function(section) {
		if (this.ldpRecord) {
			this.designer && this.designer.onTabClik(section);
			this.$profilePreference.tabs[this.ldpRecord.$uuid] = this.layoutContent.items.indexOf(section);
		}
	},
	saveSelectedTab: function(box) {
		if (this.layoutContent && this.ldpRecord) {
			var tab = this.layoutContent.getOpenedTab();
			this.saveLandingPreferences();
		}
	},
	createNewTab: function() {
		var tab = this.layoutContent.loadChildItem(null, {
			$category: "section",
			$isEmptyVisible: true,
			$title: syra_localizer.getDefaultTabTitle(this),
		});
		if (this.addTabBtn) {
			this.layoutContent.ensureTabsSlot().appendChild(this.addTabBtn.link);
		}
		tab.titleBtn.link.click();
		return tab;
	},
	_addVignette: function() {
		var self = this;
		var $url = syra_url.formatMenuUrl(self.ldpRecord, self.ldpRecord.ldp_getAddLink(), self.ldpRecord);
		syra_over.openModal(self, {
			article: self,
			$url: $url,
			$isOkHidden: true,
			onValidate: function(searchPage) {
				var sels = searchPage.selectedVignettes;
				if (sels) {
					var sendBag = syra_form.getSendBag(self.ldpRecord);
					sendBag.vignettes = sendBag.vignettes || [];
					var $serverIndex = self.ldpRecord.dataset.vignettes.length + sendBag.vignettes.length;
					for (var ii = 0, jj = sels.length; ii < jj; ii++) {
						var sel = sels[ii];
						sendBag.vignettes.push({
							$uuid: helpers.uuid.generate(),
							$index: $serverIndex++,
							vignette: {
								$uuid: sel.vignette.$key || sel.vignette.$uuid,
								title: sel.vignette.title
							},
							endpoint: {
								$uuid: sel.endpoint.$uuid
							}
						});
					}
					self.updateWorkingCopy(sendBag, self.ldpRecord, true);
				}
			}
		});
	},
	switchDesigner: function(open) {
		if (open) {
			this.designer = new Designer();
			this.designer.openDesigner(this);
		} else {
			this.designer && this.designer.dispose();
			delete this.designer;
		}
	},
	dispose: function() {
		this.saveSelectedTab();
		delete syra_site.landingPage;
		DesktopPage.prototype.dispose.call(this);
	}
});
