// syracuse-ui/lib/page/landing/landingPageChild.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var DesktopPage = require('syracuse-ui/lib/page/desktopPage').DesktopPage;

function LandingPageChild() {}

exports.LandingPageChild = helpers.defineClass(LandingPageChild, DesktopPage, {
	getAddVignetteDefinition: function() {
		var vignettes = this.masterRecord.$prototype.$properties.vignettes;
		return vignettes && vignettes.$links && vignettes.$links.$select;
	},
	appendHeader: function() {
		DesktopPage.prototype.appendHeader.call(this);
		var slot = document.createElement("div");
		slot.className = "s-ldp-head-btns";
		this.addVignetteBtn = syra_menus.button.add({
			parent: this,
			slot: slot,
			text: syra_local.ldpAddVignette,
			css: "s-ldp-btn",
			iconOnly: true,
			fontIcon: "create",
			isHidden: true,
			shortCutTip: "ESC SHIFT N",
			btnclick: function() {
				syra_site.landingPageMaster.addVignette(this.parent);
			}
		});

		this.refreshBtn = syra_menus.button.add({
			parent: this,
			slot: slot,
			text: syra_local.ldpRefresh,
			css: "s-ldp-btn",
			iconOnly: true,
			fontIcon: "refresh",
			shortCutTip: "ESC F5",
			btnclick: function() {
				var items = this.parent.layoutContent.getFields();
				for (var ii = 0, jj = items.length; ii < jj; ii++) {
					items[ii].refresh();
				}
			}
		});


		if (this.isDesignModeEnabled) {
			this.designBtn = syra_menus.button.add({
				parent: this,
				slot: slot,
				text: syra_local.ldpOpenDesigner,
				css: "s-ldp-btn",
				iconOnly: true,
				fontIcon: "settings",
				isHidden: true,
				btnclick: function() {
					this.parent.designer.toggleDesigner();
				}
			});
			syra_menus.button.hide(this.refreshBtn, true);
		}
		this.header.appendChild(slot);
	},
	ensureDefaultArticle: function($article, $prototype) {
		$article = $article || {};
		$article.$layout = $article.$layout || {};
		$article.$layout.$layoutType = "tabs";
		$article.$layout.$items = $article.$layout.$items || [];
		return $article;
	},
	_createNewTab: function() {
		var tab = this.layoutContent.loadChildItem(null, {
			$category: "section",
			$isEmptyVisible: true,
			$title: syra_site.localizer.getDefaultTabTitle(this),
		});
		tab.openBox(true);
		return tab;
	},
	onOpenerChildBoxClick: function(box) {
		if (box && box.isTabLayout) {
			this.designer && this.designer.onTabClik(box);
			syra_site.landingPageMaster.preferences.tabs = syra_site.landingPageMaster.preferences.tabs || {};
			syra_site.landingPageMaster.preferences.tabs[this.masterRecord.$uuid] = box.layoutParent.items.indexOf(box);
			syra_site.landingPageMaster.savePreferences();
		}
		return true;
	},
	removeDeletedVignettes: function($vignettes) {
		var $binds = Object.keys(this.$prototype.$properties);
		for (var ii = 0, jj = $binds.length; ii < jj; ii++) {
			var $bind = $binds[ii];
			if (!$vignettes[$bind]) {
				var fields = this.boundFields[$bind];
				if (fields) {
					for (var mm = 0, kk = fields.length; mm < kk; mm++) {
						var vignette = fields[mm];
						vignette.layoutParent.removeItem(vignette, true, true);
						delete this.$prototype.$properties[$bind];
					}
				}
				delete this.boundFields[$bind];
			}
		}
	},
	applyVignettesChange: function($vignettes) {
		this.removeDeletedVignettes($vignettes);
		var $binds = Object.keys($vignettes);
		for (var ii = 0, jj = $binds.length; ii < jj; ii++) {
			var $bind = $binds[ii];
			var $vignette = $vignettes[$bind];
			var $field = this.$prototype.$properties[$bind];
			if (!$field) {
				var tab = this.layoutContent.getOpenedTab() || this._createNewTab();
				this.$prototype.$properties[$bind] = $vignette;
				tab.layoutContent.createChildItem({
					$bind: $bind
				});
			}
		}
		if (this.designer) {
			this.designer.endArticleUpdate(null, null, false);
		}
	},
	loadBox: function(initData) {
		this.isMenuBarDisabled = true;
		this.mainPageDesignerAccess = "unknow";
		this.masterRecord = this.$prototype.masterRecord;
		delete this.$prototype.masterRecord;
		this.isDesignModeEnabled = syra_site.landingPageMaster.getSaveDesignLink(this.masterRecord) != null;
		this.isAutoInsertFieldDisabled = true;
		this.isLandingPage = true;
		this.$skin = this.$item.$skin || "s-ldp";
		this.$skinMenu = "s-ldp-nav";
		this.$defaultSkinBlock = "s-ldp-h2";
		this.$defaultSkinSection = "s-ldp-h1";
		this._defaultTitle = syra_local.ldpTitle;
		syra_site.landingPageMaster.preferences.tabs = syra_site.landingPageMaster.preferences.tabs || {};
		this.$item.$layout.$tabIndex = syra_site.landingPageMaster.preferences.tabs[this.masterRecord.$uuid] || 0;
		DesktopPage.prototype.loadBox.call(this, initData);
		if (this.isDesignModeEnabled) {
			this.addTabBtn = syra_menus.button.add({
				parent: this,
				slot: this.layoutContent.ensureTabsSlot(),
				text: "+",
				title: syra_local.ldpAddTab,
				css: "s-ldp-add-tab",
				shortCutTip: "ESC N",
				btnclick: function() {
					var page = this.parent;
					page._createNewTab();
					page.designer.endArticleUpdate();
				}
			});
			this.layoutContent.tabStop = this.addTabBtn.link;
			syra_menus.button.hide(this.refreshBtn, false);
			syra_menus.button.hide(this.designBtn, false);
		}
		this.isDesignModeEnabled && syra_site.switchItemDesigner(this, true);
		if (this.isDesignModeEnabled && this.getAddVignetteDefinition()) {
			syra_menus.button.hide(this.addVignetteBtn, false);
		}
	}
});
