// syracuse-ui/lib/page/landing/landingPageChild.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var LandingPageDesigner = require("syracuse-ui/lib/page/landing/designer/landingPageDesigner").LandingPageDesigner;
var DesktopPage = require('syracuse-ui/lib/page/desktopPage').DesktopPage;

function LandingPageChild() {}

exports.LandingPageChild = helpers.defineClass(LandingPageChild, DesktopPage, {
	designArticle: function(open) {
		if (open) {
			this.designer = new LandingPageDesigner();
			this.designer.openDesigner(this);
		} else {
			if (this.designer) {
				this.designer.dispose();
			}
			this.designer = null;
			this.resizeArticle();
		}
	},
	getAddVignetteDefinition: function() {
		var vignettes = this.masterRecord.$prototype.$properties.vignettes;
		return vignettes && vignettes.$links && vignettes.$links.$select;
	},
	appendHeader: function() {
		DesktopPage.prototype.appendHeader.call(this);
		var slot = document.createElement("div");
		slot.className = "s-ldp-head-btns";
		this.addVignetteBtn = syra_menus.addIconButton(syra_local.ldpAddVignette, "s-ldp-btns", "onAddVignetteClick", "ESC SHIFT N", "create");
		this.addVignetteBtn.style.display = "none";
		slot.appendChild(this.addVignetteBtn);
		this.refreshPicker = syra_menus.addIconButton(syra_local.ldpRefresh, "s-ldp-btns", "onRefreshPageClick", "ESC F5", "refresh");
		slot.appendChild(this.refreshPicker);
		if (this.isDesignModeEnabled) {
			this.designBtn = syra_menus.addIconButton(syra_local.ldpOpenDesigner, "s-ldp-btns", "openDesigner", null, "settings");
			this.designBtn.style.display = this.refreshPicker.style.display = "none";
			slot.appendChild(this.designBtn);
		}
		slot.syrainout = slot.syraItem = this.id;
		this.header.appendChild(slot);

	},
	onItemInOut: function(onEnter) {
		this.addVignetteBtn && syra_site.dom.toggleClass(this.addVignetteBtn, "s-enter", onEnter);
		this.refreshPicker && syra_site.dom.toggleClass(this.refreshPicker, "s-enter", onEnter);
		this.designBtn && syra_site.dom.toggleClass(this.designBtn, "s-enter", onEnter);
	},
	onRefreshPageClick: function() {
		var items = this.layoutContent.getFields();
		for (var ii = 0, jj = items.length; ii < jj; ii++) {
			items[ii].refresh();
		}
	},
	onAddVignetteClick: function() {
		syra_site.landingPageMaster.addVignette(this);
	},
	onAddTabClick: function() {
		this._createNewTab();
		this.designer.endArticleUpdate();
	},
	openDesigner: function() {
		this.designer.toggleDesigner();
	},

	ensureDefaultArticle: function($article, $prototype) {
		$article = $article || {};
		$article.$layout = $article.$layout || {};
		$article.$layout.$layoutType = "tabs";
		$article.$layout.$items = $article.$layout.$items || [];
		return $article;
	},
	_createNewTab: function() {
		var tab = this.layoutContent.loadChildItem(null, {
			$category: "section",
			$isEmptyVisible: true,
			$title: syra_site.localizer.getDefaultTabTitle(this),
		});
		tab.openBox(true);
		return tab;
	},
	onOpenerChildBoxClick: function(box, link) {
		if (box) {
			if (box.isTabLayout) {
				if (this.designer) {
					this.designer.onTabClik(box);
				}
				syra_site.landingPageMaster.preferences.tabs = syra_site.landingPageMaster.preferences.tabs || {};
				syra_site.landingPageMaster.preferences.tabs[this.masterRecord.$uuid] = box.layoutParent.getChildIndex(box);
				syra_site.landingPageMaster.savePreferences();
			}
		}
		return true;
	},
	removeDeletedVignettes: function($vignettes) {
		var $binds = Object.keys(this.$prototype.$properties);
		for (var ii = 0, jj = $binds.length; ii < jj; ii++) {
			var $bind = $binds[ii];
			if (!$vignettes[$bind]) {
				var fields = this.boundFields[$bind];
				if (fields) {
					for (var mm = 0, kk = fields.length; mm < kk; mm++) {
						var vignette = fields[mm];
						vignette.layoutParent.removeItem(vignette, true, true);
						delete this.$prototype.$properties[$bind];
					}
				}
				delete this.boundFields[$bind];
			}
		}
	},
	applyVignettesChange: function($vignettes) {
		this.removeDeletedVignettes($vignettes);
		var $binds = Object.keys($vignettes);
		for (var ii = 0, jj = $binds.length; ii < jj; ii++) {
			var $bind = $binds[ii];
			var $vignette = $vignettes[$bind];
			var $field = this.$prototype.$properties[$bind];
			if (!$field) {
				var tab = this.layoutContent.getOpenedTab() || this._createNewTab();
				this.$prototype.$properties[$bind] = $vignette;
				tab.layoutContent.createChildItem({
					$bind: $bind
				});
			}
		}
		if (this.designer) {
			this.designer.endArticleUpdate(null, null, false);
		}
	},
	getPageSlot: function() {
		return syra_site.body;
	},
	loadBox: function(initData) {
		this.isSiteRegisterDisabled = true;
		this.isMenuBarDisabled = true;
		this.mainPageDesignerAccess = "unknow";
		this.masterRecord = this.$prototype.masterRecord;
		delete this.$prototype.masterRecord;
		this.isDesignModeEnabled = syra_site.landingPageMaster.getSaveDesignLink(this.masterRecord) != null;
		this.isAutoInsertFieldDisabled = true;
		this.isLandingPage = true;
		this.$skin = this.$item.$skin || "s-ldp";
		this.$skinMenu = "s-ldp-nav";
		this.$defaultSkinBlock = "s-ldp-h2";
		this.$defaultSkinSection = "s-ldp-h1";
		this._defaultTitle = syra_local.ldpTitle;
		syra_site.landingPageMaster.preferences.tabs = syra_site.landingPageMaster.preferences.tabs || {};
		this.$item.$layout.$tabIndex = syra_site.landingPageMaster.preferences.tabs[this.masterRecord.$uuid] || 0;
		DesktopPage.prototype.loadBox.call(this, initData);
		if (this.isDesignModeEnabled) {
			this.addTabBtn = syra_menus.addTextButton("+", "s-ldp-add-tab", "onAddTabClick", "ESC N");
			this.addTabBtn.syraItem = this.id;
			this.addTabBtn.title = syra_local.ldpAddTab;
			this.layoutContent.tabStop = this.layoutContent._tabs.appendChild(this.addTabBtn);
			this.designBtn.style.display = this.refreshPicker.style.display = "";
		}
		this.isDesignModeEnabled && this.designArticle(true);
		if (this.isDesignModeEnabled && this.getAddVignetteDefinition()) {
			this.addVignetteBtn.style.display = "";
		}
	}
});
