// syracuse-tablet/html/js/multiselection/downloadController.js
"use strict";

var Base = require("syracuse-tablet/html/js/controllers/sdataController").SdataController;
var utils = require('syracuse-tablet/html/js/helpers/utils');
var modalSyncOffline = require('syracuse-tablet/html/js/ui/modals/modalSyncOffline');

/**
 * parent can be null (formController)
 */
exports.DownloadController = utils.defineClass(
	function DownloadController(queryPage) {
		if (queryPage == null || queryPage.isFacet == null || !queryPage.isFacet("$query")) {
			throw new Error("Sdata query page expected");
		}
		Base.call(this, queryPage.pageData.dataset, null, queryPage.isWorkingCopyDataUrl());
		this.page = queryPage;
	},
	Base, {

		destroy: function() {
			Base.prototype.destroy.call(this);
			this.page == null;
		},
		/**
		 * Caller queryPage
		 * Launch download process
		 */
		multiSelTriggerDownload: function() {

			var syncRootData = {
				pageProto: this.page.getPrototype(),
				rowsData: []
			};
			var selectedData = this.page.multiSelectionController.getSelectedData();
			selectedData.forEach(function(selectedItem) {
				var rowDataSet = selectedItem.rowData;
				syncRootData.rowsData.push(rowDataSet.json);
			});

			var modal = new modalSyncOffline.Modal(syncRootData);
			modal.show().then(function(result) {});
		}
	});
