// syracuse-tablet/html/js/ui/datetimeSelectors/dateSelector.js
"use strict";
var utils = require('syracuse-tablet/html/js/helpers/utils');
var uiUtils = require('syracuse-tablet/html/js/ui/uiUtils');
var modal = require('syracuse-tablet/html/js/ui/modal');
var _dateApi = require('syracuse-tablet/html/js/helpers/types/date');
var MonthChoice = require('syracuse-tablet/html/js/ui/datetimeSelectors/monthChoice').MonthChoice;
var YearChoice = require('syracuse-tablet/html/js/ui/datetimeSelectors/yearChoice').YearChoice;

var _Klass = utils.defineClass(

	function() {

	}, null, {
		onClick: function(evt) {
			var self = this,
				close;
			while (evt.target) {
				if (evt.target.getAttribute("data-s-syraOnClick")) {
					close = this[evt.target.getAttribute("data-s-syraOnClick")](evt.target, evt);
					if (close) {
						self.$$modalElmt.modal("hide");
					}
					return;
				} else {
					evt.target = evt.target.parentNode;
				}
			}
		},
		onTodayClick: function() {
			this._currentDate = _dateApi.today();
			this.field.setFieldValue(_dateApi.fromInternalValue(this._currentDate._value).toString());
			return true;
		},
		onDayClick: function(picker) {
			this.field.setFieldValue(_dateApi.fromInternalValue(picker.getAttribute("data-s-syraValue")).toString());
			return true;
		},
		onQuickItemClick: function(picker, event) {
			//TODO
			/*
			var isBegin = syra_site.dom.getChildIndex(event.target) == 0;
			this._currentDate = this._currentDate[(isBegin ? "begOf" : "endOf") + picker.getAttribute("data-s-syraInterval")]((picker.getAttribute("data-s-syraInterval") == "Week") ? 1 : undefined);
			this._drawBody();
			this.quickList.style.display = "none";
			*/
		},
		onChangePeriode: function(picker) {
			if (picker.getAttribute("data-s-syraPeriod") == "month") {
				this._currentDate = this._currentDate.addMonths(picker.getAttribute("data-s-syraIsPrev") ? -1 : 1);
			} else if (picker.getAttribute("data-s-syraPeriod") == "year") {
				this._currentDate = this._currentDate.addYears(picker.getAttribute("data-s-syraIsPrev") ? -1 : 1);
			} else {
				this._currentDate = this._currentDate.addDays(picker.getAttribute("data-s-syraIsPrev") ? -7 : 7);
			}
			this._drawBody();
		},
		onQuickClick: function() {
			modal.info("Date time selector", "implementation in progress");
			return;
			if (!this.quickList) {
				this.quickList = document.createElement("nav");
				this.quickList.style.display = "none";
				this.quickList.className = "s-calendar-quick-list";
				var intervals = ["Year", "Quarter", "Month", "Week"];
				for (var ii = 0, jj = intervals.length; ii < jj; ii++) {
					var row = document.createElement("div");
					row.className = "s-calendar-quick";
					row.setAttribute("data-s-syraInterval", intervals[ii]);
					row.setAttribute("data-s-syraOnClick", "onQuickItemClick");
					row.innerHTML = "<a class='s-calendar-quick-link'>" + this.field.localize.fdpBegin + "</a>/<a class='s-calendar-quick-link'>" + this.field.localize.fdpEnd + "</a>" + this.field.localize["fdpIntervalOf" + intervals[ii]];
					this.quickList.appendChild(row);
				}
				this.domItem.appendChild(this.quickList);
			}
			if (this.quickList.style.display == "") {
				this.quickList.style.display = "none";
			} else {
				this.quickList.style.display = "";
				$(this.quickList).position({
					my: "left bottom",
					at: "right bottom",
					of: $(this._quick)
				});
			}
		},
		onMonthClick: function() {
			var self = this;

			// hide current view
			self.domItem.style.display = "none";

			if (!self._month) {
				self._month = new MonthChoice();
				//self._month.create(self.field, self.domItem, self._currentDate, function(newDate) {
				self._month.create(self.parentSlot, self._currentDate, function(newDate) {
					self._currentDate = newDate;
					setTimeout(function() {
						self._drawBody();
						self.onMonthClick();
					}, 200);
					return true;
				});
				self._month.toggle(true);
			} else {
				self._month.toggle(false);
				self._month.destroy();
				delete self._month;
				self.domItem.style.display = "";
			}
		},
		onYearClick: function() {
			//modal.info("Date time selector", "implementation in progress");
			//return;
			//TODO
			var self = this;

			// hide current view
			self.domItem.style.display = "none";

			if (!self._year) {
				self._year = new YearChoice();
				self._year.create(self.parentSlot, self._currentDate, function(newDate) {
					self._currentDate = newDate;
					setTimeout(function() {
						self._drawBody();
						self.onYearClick();
					}, 200);
					return true;
				});
				self._year.toggle(true);
			} else {
				self._year.toggle(false);
				self._year.destroy();
				delete self._year;
				self.domItem.style.display = "";
			}
		},
		setContainer: function($$modalElmt) {
			this.$$modalElmt = $$modalElmt;
			this.parentSlot = this.$$modalElmt.find(".modal-body")[0];
		},
		create: function(field) {
			this.domItem = document.createElement("div");
			this.domItem.className = "s-calendar";

			var value = field.getValueForSelector() || "";

			this._selectedDate = value ? _dateApi.parse(value) : _dateApi.today();
			this._currentDate = _dateApi.fromInternalValue(this._selectedDate._value);

			this.field = field;
			this._table = document.createElement("table");
			this._table.setAttribute("cellspacing", 0);
			this._table.className = "s-calendar-content";
			this._table.appendChild(this._appendHead());
			this._table.appendChild(this.body = document.createElement("tbody"));
			this._table.appendChild(this._appendFoot());
			this.domItem.appendChild(this._table);

			this._drawBody();

			return this.domItem;
		},
		_appendHead: function() {
			var head = document.createElement("thead");
			var row = document.createElement("tr");
			var cell = document.createElement("th");
			cell.setAttribute("colspan", 8);
			cell.className = "s-calendar-month-year";

			var slot = document.createElement("div");
			slot.className = "s-calendar-month-year-slot";
			cell.appendChild(slot);

			// prev year link
			var link = document.createElement("a");
			link.setAttribute("data-s-syraOnClick", "onChangePeriode");
			link.setAttribute("data-s-syraIsPrev", true);
			link.setAttribute("data-s-syraPeriod", "year");
			link.className = "s-calendar-prev s-calendar-prev-year glyphicon glyphicon-fast-backward";
			slot.appendChild(link);

			// prev month link
			link = document.createElement("a");
			link.setAttribute("data-s-syraOnClick", "onChangePeriode");
			link.setAttribute("data-s-syraIsPrev", true);
			link.setAttribute("data-s-syraPeriod", "month");
			link.className = "s-calendar-prev s-calendar-prev-month glyphicon glyphicon-step-backward";
			slot.appendChild(link);

			this._monthLink = document.createElement("a");
			this._monthLink.className = "s-month s-calendar-month-year-link";
			slot.appendChild(this._monthLink).setAttribute("data-s-syraOnClick", "onMonthClick");

			this._yearLink = document.createElement("a");
			this._yearLink.className = "s-year s-calendar-month-year-link";
			slot.appendChild(this._yearLink).setAttribute("data-s-syraOnClick", "onYearClick");

			// next month link
			link = document.createElement("a");
			link.setAttribute("data-s-syraOnClick", "onChangePeriode");
			link.setAttribute("data-s-syraPeriod", "month");
			link.className = "s-calendar-next s-calendar-next-month glyphicon glyphicon-step-forward";
			slot.appendChild(link);

			// next year link
			link = document.createElement("a");
			link.setAttribute("data-s-syraOnClick", "onChangePeriode");
			link.setAttribute("data-s-syraPeriod", "year");
			link.className = "s-calendar-next s-calendar-next-year glyphicon glyphicon-fast-forward";
			slot.appendChild(link);

			row.appendChild(cell);
			head.appendChild(row);

			row = document.createElement("tr");
			var cell = document.createElement("th");
			cell.className = "s-calendar-week-day";
			row.appendChild(cell);

			var days = [1, 2, 3, 4, 5, 6, 0];
			for (var ii = 0, jj = days.length; ii < jj; ii++) {
				var cell = document.createElement("th");
				cell.className = "s-calendar-week-day";
				cell.title = _dateApi.dayName(days[ii]);
				cell.textContent = _dateApi.dayName(days[ii], true);
				row.appendChild(cell);
			}
			head.appendChild(row);
			return head;
		},
		_appendFoot: function() {
			var row = document.createElement("tr");
			var cell = document.createElement("td");
			cell.setAttribute("colspan", 3);
			cell.className = "s-calendar-foot-week";

			var link = document.createElement("a");
			link.setAttribute("data-s-syraOnClick", "onChangePeriode");
			link.setAttribute("data-s-syraPeriod", "week");
			link.className = "s-calendar-prev s-calendar-prev-week glyphicon glyphicon-chevron-left";
			link.setAttribute("data-s-syraIsPrev", true);
			cell.appendChild(link);

			var label = document.createElement("label");
			label.className = "s-calendar-foot-week-title";
			label.textContent = "Week";
			//TODO
			//label.textContent = this.field.localize.fdpWeek;
			cell.appendChild(label);

			this._weekNumber = document.createElement("label");
			this._weekNumber.className = "s-calendar-foot-week-title-num";
			cell.appendChild(this._weekNumber);

			link = document.createElement("a");
			link.setAttribute("data-s-syraOnClick", "onChangePeriode");
			link.setAttribute("data-s-syraPeriod", "week");
			link.className = "s-calendar-next s-calendar-next-week glyphicon glyphicon-chevron-right";
			link.setAttribute("data-s-syraIsPrev", false);
			cell.appendChild(link);
			row.appendChild(cell);

			cell = document.createElement("td");
			cell.setAttribute("colspan", 3);
			cell.className = "s-calendar-foot-today";
			link = document.createElement("a");
			link.className = "s-calendar-today-link";
			link.setAttribute("data-s-syraOnClick", "onTodayClick");

			//TODO
			//link.textContent = this.field.localize.fdpToday;
			link.textContent = "Today";
			cell.appendChild(link);
			row.appendChild(cell);

			cell = document.createElement("td");
			cell.setAttribute("colspan", 2);
			cell.className = "s-calendar-foot-more";
			this._quick = document.createElement("a");
			this._quick.className = "s-calendar-quick-btn glyphicon glyphicon-plus";
			this._quick.setAttribute("data-s-syraOnClick", "onQuickClick");
			cell.appendChild(this._quick);
			row.appendChild(cell);

			var foot = document.createElement("tfoot");
			foot.appendChild(row);
			return foot;
		},
		_drawBody: function() {
			//TODO
			//var curMonth = this._currentDate.month
			var curMonth = this._currentDate.month || 9;
			var month = _dateApi.monthName(curMonth);

			this._monthLink.textContent = month;
			this._yearLink.textContent = this._currentDate.year || 2014;
			//TODO
			//this._yearLink.textContent = this._currentDate.year;
			uiUtils.empty(this.body);

			var curDate = _dateApi.fromInternalValue(this._currentDate._value);
			var begOfMonth = curDate = curDate.begOfMonth();
			curDate = curDate.begOfWeek(1);

			for (var weekRow = 0; weekRow < 6; weekRow++) {
				var row = document.createElement("tr");
				var weekDay = (weekRow == 0) ? begOfMonth : curDate;
				var cell = document.createElement("td");
				cell.className = "s-calendar-week-num";
				row.appendChild(cell);
				cell.textContent = weekDay.week;
				for (var day = 0; day < 7; day++) {
					var cell = document.createElement("td");
					cell.className = "s-calendar-day-link";
					cell.setAttribute("data-s-syraOnClick", "onDayClick");
					cell.setAttribute("data-s-syraValue", curDate._value);
					var link = document.createElement("a");
					link.textContent = curDate.day;
					if (curMonth != curDate.month) {
						cell.className += " s-calendar-other-month";
					}
					if (this._currentDate.equals(curDate)) {
						cell.className += " s-calendar-select";
						link.className = "s-calendar-select";
					}
					cell.appendChild(link);
					row.appendChild(cell);
					curDate = curDate.addDays(1);
				}
				this.body.appendChild(row);
			}
			this._weekNumber.textContent = this._currentDate.week;
		},
		bindEvents: function(bind) {
			var self = this;
			var $$calendar = $(".s-calendar");
			if (bind) {
				// TODO improve
				// self.body, self.domItem and other dom variables are not tied to the dom and have to be reset
				self.domItem = $$calendar[0];
				self.body = $$calendar.find("tbody")[0];
				self._monthLink = $$calendar.find(".s-month")[0];
				self._yearLink = $$calendar.find(".s-year")[0];
				self._weekNumber = $$calendar.find(".s-calendar-foot-week-title-num")[0];

				$$calendar.delegate("td, a", "click", function(evt) {
					self.onClick(evt);
					evt.preventDefault();
					evt.stopPropagation();
				});
			} else {
				$$calendar.undelegate();

				// unbind monthchoice and yearchoice components if necessary
				if (self._month) {
					self._month.bindEvents(false);
				}
				if (self._year) {
					self._year.bindEvents(false);
				}
			}
		},
		destroy: function() {
			//TODO to complete
			this.bindEvents(false);
			this._currentDate = this.field = this._currentDate = this.quickList = this._month = this._year = this._selectedDate = this._monthLink = this._yearLink = this._weekNumber = this._quick = this.body = this._weekNumber = this.domItem = null;
		}
	});

exports.Klass = _Klass;
