// syracuse-tablet/html/js/ui/datetimeSelectors/selectorSwitchHandler.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var uiUtils = require("syracuse-tablet/html/js/ui/uiUtils");
var MonthChoice = require('syracuse-tablet/html/js/ui/datetimeSelectors/monthChoice').MonthChoice;
var YearChoice = require('syracuse-tablet/html/js/ui/datetimeSelectors/yearChoice').YearChoice;
var DateSelector = require('syracuse-tablet/html/js/ui/datetimeSelectors/dateSelector').Klass;
var TimeSelector = require('syracuse-tablet/html/js/ui/datetimeSelectors/timeSelector').Klass;
var DatetimeSelector = require('syracuse-tablet/html/js/ui/datetimeSelectors/datetimeSelector').Klass;

//var events = ['openYear','openMonth','openTime','selectMonth','selectYear','selectTime'];
var _Klass = utils.defineClass(

	function(field) {
		this.field = field;
	}, null, {
		onEvent: function(evt, selector) {
			this[evt](selector);
		},
		setContainer: function($$container) {
			this.$$container = $$container;
		},
		setParentSlot: function(parentSlot) {
			this.parentSlot = parentSlot;
		},
		getParentSlot: function() {
			return this.parentSlot;
		},
		onMonthClick: function(selector) {
			var self = this;
			// save date
			var date = selector._currentDate;

			// clear previous selector
			uiUtils.empty(self.parentSlot);
			selector.destroy();

			// create month selector
			var monthChoice = new MonthChoice();
			monthChoice.create(this.parentSlot, date, self);
		},
		onYearClick: function(selector) {
			var self = this;
			// save date
			var date = selector._currentDate;

			// clear previous selector
			uiUtils.empty(self.parentSlot);
			selector.destroy();

			// create month selector
			var yearChoice = new YearChoice();
			yearChoice.create(self.parentSlot, date, self);
		},
		_onOpenTime: function(selector) {

		},
		onCalendarUpdate: function(selector) {
			var self = this;

			// save date
			self._currentDate = selector._currentDate;

			// clear previous selector
			uiUtils.empty(self.parentSlot);
			selector.destroy();

			// create back date selector
			var dateSelector = new DateSelector();
			var sDom = dateSelector.create(self.field, self);
			self.parentSlot.appendChild(sDom);
			dateSelector.bindEvents(true);
		},
		_onSelectTime: function(selector) {

		},
		closeContainer: function() {
			this.$$container.modal("hide");
		}
	});

exports.Klass = _Klass;
