// syracuse-tablet/html/js/widgets/field.js
"use strict";

var utils = require("syracuse-tablet/html/js/helpers/utils");

function Field() {}

exports.Field = utils
	.defineClass(
		Field,
		null, {
			init: function($item) {
				var cssField = "sm-field";
				this.$isEditMode = $item.$isEditMode;

				// declare container
				this.layoutSlot = document.createElement("div");
				this.layoutSlot.className = $item.$css || cssField;

				// title container
				this.titleDom = document.createElement("div");
				this.titleDom.className = cssField + "-title";
				this.titleDom.textContent = $item.$title;

				// value container
				this.core = document.createElement("div");
				this.core.className = cssField + "-core";
				this.valueSlot = document.createElement("div");
				this.valueSlot.className = cssField + "-value-slot";
				if (this.$isEditMode) {
					// TODO
				} else {
					this.valueDom = document.createElement("div");
					this.valueDom.className = cssField + "value";
					this.valueSlot.appendChild(this.valueDom);
				}

				// appending
				this.core.appendChild(this.valueSlot);
				this.layoutSlot.appendChild(this.titleDom);
				this.layoutSlot.appendChild(this.core);
			},
			render: function() {
				return this.layoutSlot;
			},
			setDataValue: function(value) {
				if (this.isEditMode) {
					// TODO
				} else {
					this.valueDom.textContent = value;
				}
			},
			dispose: function() {
				this.$isEditMode = null;
				this.layoutSlot = this.titleDom = this.core = this.valueDom = this.valueSlot = null;
			}
		});

exports.create = function(data, options) {
	var o = new Field();
	o.init(data, options);
	return o;
};
