// syracuse-uml/lib/representations/uml-class-diagram.js
"use strict";
module.exports = {
	$entityName: "umlClassDiagram",
	$facets: {
		$details: {
			$layout: {
				$items: [{
					$layoutType: "row",
					$items: [{
						$items: [{
							$bind: "name"
						}, {
							$bind: "endpoint"
						}, {
							$bind: "module"
						}],
					}, {
						$items: [{
							$bind: "linksTo"
						}, {
							$bind: "linkedFrom"
						}],
					}],
				}, {
					$bind: "data"
				}],
			},
		},
		$edit: {
			$layout: {
				$items: [{
					$layoutType: "row",
					$items: [{
						$items: [{
							$bind: "name"
						}, {
							$bind: "endpoint"
						}, {
							$bind: "module"
						}],
					}, {
						$items: [{
							$bind: "linksTo"
						}],
					}],
				}, {
					$bind: "data"
				}],
			},
		},
		$query: {
			$layout: {
				$items: [{
					$category: "section",
					$layout: {
						$items: [{
							$category: "section",
							$layout: {
								$items: [{
									$bind: "$resources",
									$layout: {
										$items: [{
											$bind: "name"
										}, {
											$bind: "endpoint"
										}, {
											$bind: "module"
										}]
									}
								}]
							}
						}]
					}
				}]
			}
		},
		$lookup: {
			$copy: "$query"
		},
		$select: {
			$copy: "$lookup"
		}
	}
};
