// syracuse-ui/lib/mobile/site/mobilePanel.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Article = require("syracuse-ui/lib/common/article/article").Article;

var _localize = {
    $about: "About",
    $drafts: "Drafts",
    $historic: "Historic",
    $sessions: "Sessions",
    $bookmarks: "Bookmarks",
    $dashboards: "Dashboard",
    $bookmarkLinkTitle: "Add this page?",
    $bookmarkLinkDesc: "It will appear in your bookmark",
    $authoringLinkTitle: "Design the current page?",
    $authoringLinkDesc: "Launch the authoring panel"
};


function MobilePanel(){
}

exports.MobilePanel = helpers.defineClass(MobilePanel, Article, {
    loadBox: function(){
        var self = this;
        
        self.navItems = {
            drafts: {
                $title: _localize.$drafts
            },
            historic: {
                $title: _localize.$historic,
                $links: {
                    banks: {
                        $title: "List of banks"
                    },
                    properties: {
                        $title: "Bank properties (Boursorama)"
                    },
                    newAsset: {
                        $title: "New asset"
                    },
                    listAsset: {
                        $title: "List of assets"
                    },
                    account: {
                        $title: "Edit bank account (BNP)"
                    }
                }
            },
            bookmarks: {
                $title: _localize.$bookmarks
            },
            dashboards: {
                $title: _localize.$dashboards,
                $links: document.site.$prototype.$navigation.$dashboard
            }
        }
        
        self.$item = {
            $skin: "s-portlet",
            $layout: {
                $items: self._makeSection(["drafts", "historic", "bookmarks", "dashboards"])
            }
        };
        
        Article.prototype.loadBox.call(this);
    },
    _makeSection: function($types){
        var self = this;
        return $types.map(function($type){
            return {
                $category: "section",
                $title: self.navItems[$type].$title,
                $layout: {
                    $items: [{
                        $category: "links",
                        $isBindDisabled: true,
                        $links: self.navItems[$type].$links || {}
                    }]
                }
            }
        });
    },
});
