// syracuse-ui/lib/site/help/helpCenter.js
"use strict";
var LegalPage = require('syracuse-ui/lib/site/help/legal/legalPage').LegalPage;

var _popupSlot, _popup;
var _browserTab = {};

exports.openByShortCut = function(page, field) {
	page.externalAdapter.onFieldClickPicker({
		field: field || page,
		pickerType: "help",
		doEvent: function() {
			exports.open(page, field);
		}
	});
};

exports.open = function(page, field, $url) {
	var winId = "sageerphelp";
	if (!$url) {
		var $links = page.$prototype.$links;
		$url = $links && $links.$help && $links.$help.$url;
		if ($url) {
			if (field) {
				$url = syra_site.expressionMaker.parse(field.articleParent, $url) + '#' + field.$item.$bind;
			} else {
				$url = syra_site.expressionMaker.parse(page, $url);
			}
		}
	}
	if ($url) {
		if (syra_site.browser.getIdentity().isMSIE) {
			// Troubles with IE... to improve!
			window.open($url, "_blank");
			return;
		}
		if (!_browserTab.winRef || _browserTab.winRef.closed == undefined || _browserTab.winRef.closed) {
			_browserTab.winRef = window.open($url, winId, "width=800,height=600,resizable,scrollbars,status");
		} else
		if (_browserTab.lastUrl != $url) {
			_browserTab.winRef = window.open($url, winId, "resizable=true,scrollbars=yes,status=yes");
			if (_browserTab.winRef && _browserTab.winRef.focus) {
				_browserTab.winRef.focus();
			}
		} else
		if (_browserTab.winRef && _browserTab.winRef.focus) {
			_browserTab.winRef.focus();
		}
		_browserTab.lastUrl = $url;
	}
};




exports.openPopup = function(menuItem) {
	if (!_popupSlot) {
		_popupSlot = document.createElement("div");
		_popupSlot.className = "s-mn-popus-body";
		var $links = syra_site.$prototype.$helpLinks;
		for (var ii = 0, jj = $links.length; ii < jj; ii++) {
			syra_site.loadNewItem(_popupSlot, {
				$bind: $links[ii],
				$category: "link",
				$skin: "s-mn-link"
			});
		}
	}
	if (!_popup) {
		_popup = syra_site.dialogManager.openPopup(syra_site, {
			content: syra_site,
			slot: _popupSlot,
			position: {
				my: "left top",
				at: "left bottom",
				of: $(menuItem.domItem)
			},
			onClose: function() {
				setTimeout(function() {
					_popup = null;
				}, 200);
			}
		});
	} else {
		_popup.close();
	}
};

exports.openLegalPage = function(menuItem) {
	syra_controller.sendRequest(null, {
		method: "GET",
		$location: {
			$url: syra_site.$item.$legalPrototypeUrl
		}
	}, function(data, response, $url) {
		var page = new LegalPage();
		page.loadBox(data);
		if (page.menuBar) {
			page.menuBar.toggleBar(false);
		}
		syra_controller.sendRequest(null, {
			method: "GET",
			$location: {
				$url: syra_site.$item.$legalDataUrl
			}
		}, function(data, response, $url) {
			page.applyChange(data);
			syra_site.dialogManager.openModal(menuItem.page, {
				page: page
			});
		}, function(error) {
			var $diagnoses;
			if (error.data.indexOf("$diagnoses") != -1) {
				$diagnoses = JSON.parse(error.data).$diagnoses;
			} else {
				$diagnoses = [{
					$severity: "error",
					$message: error.data
				}];
			}
			syra_site.showDiagnoses({
				$diagnoses: $diagnoses
			});
		});

	}, function(error) {
		var $diagnoses;
		if (error.data.indexOf("$diagnoses") != -1) {
			$diagnoses = JSON.parse(error.data).$diagnoses;
		} else {
			$diagnoses = [{
				$severity: "error",
				$message: error.data
			}];
		}
		syra_site.showDiagnoses({
			$diagnoses: $diagnoses
		});
	});

};

exports.dispose = function() {
	_browserTab = _popupSlot = _popup = null;
};
