// syracuse-tablet/html/js/helpers/native/nativeImage.js
"use strict";

var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("nativeCamera");
var utils = require('syracuse-tablet/html/js/helpers/utils');
var jsutils = require('syracuse-tablet/html/js/helpers/jsutils');
var nativeTestImage = require('syracuse-tablet/html/js/helpers/native/test/nativeTestImage');
var nativeHelpers = require('syracuse-tablet/html/js/helpers/native/nativeHelpers');
var nativeExtCall = require('syracuse-tablet/html/js/helpers/native/nativeExtCall');

var _currentDeferred;

// Global object to be called by native code on pencil events
if (!(window.smImageJS)) {
	window.smImageJS = {

		/*
		 *  data:
		 *  {
		 *  	Action: "ok" || "cancelled" || "error"
		 *  	ImageData: "data:image/png;base64,iVBORw0KGgoAAAANSUhE....",
		 *  	Message: "Error message in case action=error"
		 *  }
		 */
		getPictureFromCameraFinished: function(data) {
			log && log("Got image!");
			if (_currentDeferred) {
				var def = _currentDeferred;
				_currentDeferred = null;
				log && log("Got image: " + JSON.stringify(data || {}));
				def.resolve({
					action: data.Action,
					imageData: data.ImageData
				});
			}
		}
	};
}
/**
 *
 */
exports.supports = function(capability) {
	if (capability === "camera") {
		return nativeExtCall.isSupported("smImage", "GetPictureFromCameraSupported");
	}
	return false;
};

/**
 * Options are availabe as parameter but it's not used yet!
 * options:
 * {
 * }
 */
exports.getPictureFromCamera = function(options) {
	// No options yet!
	var param = {};
	var data = JSON.stringify(param);
	_currentShareDeferred = $.Deferred();
	nativeExtCall.fireMethod("smImage", "GetPictureFromCamera", data);
	return _currentShareDeferred.promise();
};

exports.init = function(testMode) {
	if (window && window.smImage) return;
	if (!testMode) return;
	log && log("Creating native image test wrapper");
	window.smImage = nativeTestImage.create();
};
