// syracuse-tablet/html/js/helpers/native/nativeVoiceCommands.js
"use strict";

var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("nativeVoice", true);
var locale = require('syracuse-tablet/html/js/helpers/locale');

var _currentBookmarks = [];

/**
 *
 */
exports.setVoiceBookmarks = function(titles, operation) {
	operation = operation  || "set";
	if (operation === "set") {
		_currentBookmarks = titles;
	} else if (operation === "add") {
		titles.forEach(function(t) {
			_currentBookmarks.push(t);
		});
	} else if (operation === "remove") {
		titles.forEach(function(t) {
			var idx;
			while ((idx = _currentBookmarks.indexOf(t)) >= 0) {
				_currentBookmarks.splice(idx, 1);
			}
		});
	}

	var lang = locale.getCurrentLocale().toLowerCase();;
	var data = JSON.stringify({
		phrase: "bookmark",
		phraseList: _currentBookmarks,
		language: lang
	});

	log && log("setVoiceBookmarks: " + JSON.stringify(data));
	
	// Be carefull, JS name is lowercase (e.g. setPhraseList) while C# name must be uppercase (e.g. SetPhraseList)
	window && window.smNativeVoice && window.smNativeVoice.setPhraseList && window.smNativeVoice.setPhraseList(data);
};

/**
 * 
 */
exports.addVoiceBookmark = function(title) {
	exports.setVoiceBookmarks([title], "add");
}

/**
 * 
 */
exports.removeVoiceBookmark = function(title) {
	exports.setVoiceBookmarks([title], "remove");
}
