// syracuse-tablet/html/js/helpers/native/nativeSensors.js
"use strict";

var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("nativeSensors", true);
var nativeTestSensors = require('syracuse-tablet/html/js/helpers/native/wrapperTest/nativeTestSensors');
var nativeExtCall = require('syracuse-tablet/html/js/helpers/native/nativeExtCall');

var _currentDeferred;

// Global object to be called by native code on pencil events
if (!(window.smSensorsJS)) {
	window.smSensorsJS = {

		/*
		 *  data:
		 *  {
		 *  	Latitude: 1.289374,
		 *  	Longitude: 14.1234124,
		 *  	Succeeded: true/false,
		 *  	Error: unknown,timeout,denied
		 *  }
		 */
		getGPSCoordinatesFinished: function(data) {
			log && log("Got GPS");
			if (_currentDeferred) {
				var def = _currentDeferred;
				_currentDeferred = null;
				log && log("Got GPS: " + JSON.stringify(data || {}));

				def.resolve({
					latitude: data.Latitude,
					longitude: data.Longitude,
					succeeded: data.Succeeded,
					error: data.Error
				});
			}
		}
	};
}
/**
 *
 */
exports.supports = function(capability) {
	if (capability === "GPS") {
		return nativeExtCall.isSupported("smSensors", "getGPSCoordinatesSupported");
	}
	return false;
};

/**
 */
exports.getGPSCoordinates = function() {
	_currentDeferred = $.Deferred();
	nativeExtCall.fireMethod("smSensors", "getGPSCoordinates");
	return _currentDeferred.promise();
};
/**
 * Returns the test js wrapper
 */
exports.init = function(testMode) {
	if (window && window.smSensors) return;
	if (!testMode) return;
	log && log("Creating native gps test wrapper");
	window.smSensors = nativeTestSensors.create();
	return window.smSensors;
};
