// syracuse-tablet/html/js/helpers/native/nativeLiveTile.js
"use strict";

var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("nativeLiveTile", true);

var _currentDeferred;
//Global object to be called by native code on pencil events
if (!(window.smLiveTileJS)) {
	window.smLiveTileJS = {

		/*
		 *  This method will be invoked from wrapper when there was a live tile action executed.
		 *  data {
		 *  	action :"add" //"remove" "get"
		 *  	content:{ (null in case of get if not tile found)
		 *  			id:"id passed",
		 *  			title:"my customers",
		 *  			...
		 * 		}
		 *  	error:{ (if error occurs)
		 *  		no:"132001"
		 *  		text:"Not found", "unexpected error" ...
		 *  	}
		 *  }
		 */
		liveTileActionFinished: function(data) {
			if (_currentDeferred) {
				var def = _currentInputDeferred;
				_currentDeferred = null;
				log && log("Command live tile finished: " + JSON.stringify(data || {}));
				def.resolve(data);
			}
		}
	};
}
/**
 *
 */
/*
data:{
	action:"add",
	content:{
		"name":"x3.erp.GX3APP.BPCUSTOMERM.$query",
		"title":"Clients",
		"id":"x3.erp.GX3APP.BPCUSTOMERM.$query_http://pc101458:8124/mobile1/x3/erp/GX3APP/BPCUSTOMER?representation=BPCUSTOMERM.$query",
		"icon":"star",
		"sDataUrl":"http://pc101458:8124/mobile1/x3/erp/GX3APP/BPCUSTOMER?representation=BPCUSTOMERM.$query",
		"representation":"BPCUSTOMERM",
		"facet":"$query",
		"action":"$query",
		"entity":"BPCUSTOMER",
		"endpoint":"x3.erp.GX3APP",
		"applicationName":"sagesalescustomers"
	}
}
data:{
	action:"remove"
	content:{
		id:"x3.erp.GX3APP.BPCUSTOMERM.$query_http://pc101458:8124/mobile1/x3/erp/GX3APP/BPCUSTOMER?representation=BPCUSTOMERM.$query"
	}
}
data:{
	action:"get"
	content:{
		id:"x3.erp.GX3APP.BPCUSTOMERM.$query_http://pc101458:8124/mobile1/x3/erp/GX3APP/BPCUSTOMER?representation=BPCUSTOMERM.$query"
	}
}
*/
exports.actionLiveTile = function(options) {
	var data = JSON.stringify(options);
	if (window && window.smLiveTileJS && window.smLiveTileJS.actionLiveTile) {
		_currentDeferred = $.Deferred();
		window.smLiveTileJS.actionLiveTile(data);
		return _currentDeferred.promise();
	} else {
		/*test
		if(options.action=="get"){
			return $.smResolve({action:options.action,content:null})
		}
		if(options.action=="remove"){
			return $.smResolve({action:options.action,content:{id:options.content.id},error:{no:"121212",text:"plouuffff"}})
		}
		if(options.action=="add"){
			return $.smResolve({action:options.action,content:{id:options.content.id},error:{no:"121212",text:"plouuffff"}})
		}
		End test*/
		return $.smResolve(null);
	}
};
exports.supportsLiveTile = function() {
	var supported = (window && window.smLiveTileJS && window.smLiveTileJS.supportsLiveTile && window.smLiveTileJS.supportsLiveTile()) || false;
	log && log("Live tile : " + supported);
	return supported;
};
/**
 *
 */
exports.addLiveTile = function(options) {
	return exports.actionLiveTile({
		action: "add",
		content: options
	});
};

/**
 *
 */
exports.removeLiveTile = function(options) {
	return exports.actionLiveTile({
		action: "remove",
		content: {
			id: options
		}
	});
};
exports.getLiveTile = function(options) {
	return exports.actionLiveTile({
		action: "get",
		content: {
			id: options
		}
	});
};
