// syracuse-tablet/html/js/helpers/native/nativeShare.js
"use strict";

var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("nativeShare", false);
var nativeExtCall = require('syracuse-tablet/html/js/helpers/native/nativeExtCall');
var modules = require('syracuse-tablet/html/js/common/modules');
var notifications = require('syracuse-tablet/html/js/helpers/notifications');


var _ShareManager = function() {

	this._currentDeferred = null;
	this._pendingUrl = null;
	this._isReady = false;

	/*
	 *  This method will be invoked from wrapper once the pen input has finished or failed
	 *  data:
	 *  {
	 *  	action: "ok" | "cancelled",
	 *  }
	 */
	this.shareFinished = function(data) {
		if (this._currentDeferred) {
			var def = this._currentDeferred;
			this._currentDeferred = null;
			log && log("Got share: " + JSON.stringify(data || {}));
			def.resolve(data);
		}
	};

	/*
	 * Called by wrapper when the app is activated by a link
	 * data:
	 * {
	 * 	Url: url (url containing all information to go back to a specific page, sent as url option by sharePageLink)
	 *  Title: Title of url (sent as title option by sharePageLink)
	 * }
	 */
	this.handleLink = function(data) {
		log && log("handleLink: " + JSON.stringify(data, null, 2));
		var url = data && data.Url;
		if (url) {
			this._pendingUrl = url;
			this._gotoUrl();
		}
	};

	this.notifNavigatedToPage = function(page) {
		var name = page && page.pageName;
		log && log("DeepLink: navigated to: " + name)
		if (name != null && name != "logout" && name != "login") {
			this._isReady = true;
			this._gotoUrl();
		}
	};
	/**
	 * Invoked only when ready that means that user logged in or logged out
	 */
	this._gotoUrl = function() {
		if (!this._isReady || !this._pendingUrl) {
			return;
		}
		var url = this._pendingUrl;
		this._pendingUrl = null;
		modules.get("navHelper").gotoUrl(url);
	};

	// End  of initialization
	notifications.subscribe(this, ["sm.navigated.to.page"]);
};

// Global object to be called by native code on pencil events
if (!(window.smShareJS)) {
	window.smShareJS = new _ShareManager();
}
/**
 *
 */
exports.supports = function(capability) {
	if (capability === "share") {
		return nativeExtCall.isSupported("smShare", "sharePageLinkSupported");
	}
	return $.smResolve(false);
};

/**
 * options:
 * {	
      pageData: {
		title: "Customers" // title that will be shown when page is shared
		url: url that is embedded in sent message and will be passed to handleLink method once the user clicks on the message 
		
		--- optional for now:
		
		id:"..."
		sdataParameters:"..."
		endPoint: "x3.erp.SEEDAMBAS"
		htmlRoot: "http://pc101458:8124/syracuse-tablet/html"
		language: "en-US"
		name: "x3.erp.GX3APP.BPCUSTOMERM.$query"
		sDataUrl: "http://pc101458:8124/mobile1/x3/erp/GX3APP/BPCUSTOMER?representation=BPCUSTOMERM.$query",
		
		DeepLink: {
			Title,
			Url,
		}
	  }
 * }
 */
exports.sharePageLink = function(options) {
	// DeepLink.Url must contain all information required to allow the webapp to navigate back to the current page including it's
	// eventual state (like filters)
	// DeepLink data structure will be shared by other similar functions
	// DeepLink: {
	//		Title: 
	//		Url: 
	//	}

	var param = {
		Title: options.pageData.Title,
		Description: options.pageData.Description,
		DeepLink: options.pageData.DeepLink
	};

	var data = JSON.stringify(param);
	_ShareManager._currentDeferred = $.Deferred();
	nativeExtCall.fireMethod("smShare", "sharePageLink", data);
	return _ShareManager._currentDeferred.promise();
};
/**
 */
exports.init = function(testMode) {};
