// syracuse-tablet/html/js/helpers/native/nativeShare.js
"use strict";

var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("nativeShare");
var nativeTestShare = require('syracuse-tablet/html/js/helpers/native/wrapperTest/nativeTestShare');
var nativeExtCall = require('syracuse-tablet/html/js/helpers/native/nativeExtCall');

var _ShareManager = function() {

	this.enabled = false;
	this.pendingLinks = null;
	this._currentDeferred = null;

	/*
	 *  This method will be invoked from wrapper once the pen input has finished or failed
	 *  data:
	 *  {
	 *  	action: "ok" | "cancelled",
	 *  }
	 */
	this.shareFinished = function(data) {
		if (this._currentDeferred) {
			var def = this._currentDeferred;
			this._currentDeferred = null;
			log && log("Got share: " + JSON.stringify(data || {}));
			def.resolve(data);
		}
	}
};

// Global object to be called by native code on pencil events
if (!(window.smShareJS)) {
	window.smShareJS = new _ShareManager();
}
/**
 *
 */
exports.supports = function(capability) {
	if (capability === "share") {
		return nativeExtCall.isSupported("smShare", "sharePageLinkSupported");
	}
	return $.smResolve(false);
};

/**
 * options:
 * {	
      pageData: {
		title: "Customers" // title that will be shown when page is shared
		url: url that is embedded in sent message and will be passed to handleLink method once the user clicks on the message 
		
		--- optional for now:
		
		id:"..."
		sdataParameters:"..."
		endPoint: "x3.erp.SEEDAMBAS"
		htmlRoot: "http://pc101458:8124/syracuse-tablet/html"
		language: "en-US"
		name: "x3.erp.GX3APP.BPCUSTOMERM.$query"
		sDataUrl: "http://pc101458:8124/mobile1/x3/erp/GX3APP/BPCUSTOMER?representation=BPCUSTOMERM.$query",
		
		DeepLink: {
			Title,
			Url,
		}
	  }
 * }
 */
exports.sharePageLink = function(options) {
	// DeepLink.Url must contain all information required to allow the webapp to navigate back to the current page including it's
	// eventual state (like filters)
	// DeepLink data structure will be shared by other similar functions
	// DeepLink: {
	//		Title: 
	//		Url: 
	//	}

	var param = {
		Title: options.pageData.Title,
		Description: options.pageData.Description,
		DeepLink: options.pageData.DeepLink
	};

	var data = JSON.stringify(param);
	_ShareManager._currentDeferred = $.Deferred();
	nativeExtCall.fireMethod("smShare", "sharePageLink", data);
	return _ShareManager._currentDeferred.promise();
};
/**
 * Returns the test js wrapper
 */
exports.init = function(testMode) {
	if (window && window.smShare) return;
	if (!testMode) return;
	log && log("Creating native share test wrapper");
	window.smShare = nativeTestShare.create();
	return window.smShare;
};
