// syracuse-tablet/html/js/helpers/native/nativeShare.js
"use strict";

var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("nativeShare", true);

var _currentShareDeferred;

// Global object to be called by native code on pencil events
if (!(window.smShareJS)) {
	window.smShareJS = {

		/*
		 *  This method will be invoked from wrapper once the pen input has finished or failed
		 *  data:
		 *  {
		 *  	action: "ok" | "cancelled",
		 *  }
		 */
		shareFinished: function(data) {
			if (_currentShareDeferred) {
				var def = _currentShareDeferred;
				_currentShareDeferred = null;
				log && log("Got share: " + JSON.stringify(data || {}));
				def.resolve(data);
			}
		}
	};
}
/**
 *
 */
exports.supportsShare = function() {
	var supported = (window && window.smShare && window.smShare.supportsShare && window.smShare.supportsShare()) || false;
	log && log("Share support: " + supported);
	return supported;
};

/**
 * options:
 * {	
		endPoint: "x3.erp.SEEDAMBAS"
		htmlRoot: "http://pc101458:8124/syracuse-tablet/html"
		language: "en-US"
		name: "x3.erp.GX3APP.BPCUSTOMERM.$query"
		sDataUrl: "http://pc101458:8124/mobile1/x3/erp/GX3APP/BPCUSTOMER?representation=BPCUSTOMERM.$query"
		title: "Customers"
 * }
 */
exports.sharePageLink = function(options) {
	var data = JSON.stringify(options);
	log && log("sharePageLink: " + data);
	// Be carefull, JS name is lowercase (e.g. setPhraseList) while C# name must be uppercase (e.g. SetPhraseList)
	if (window && window.smShare && window.smShare.sharePageLink) {
		_currentShareDeferred = $.Deferred();
		window.smShare.sharePageLink(data);
		return _currentShareDeferred.promise();
	} else {
		return $.smResolve(null);
	}
};
