// syracuse-tablet/html/js/helpers/native/nativeImage.js
"use strict";

var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("nativeCamera");
var nativeTestImage = require('syracuse-tablet/html/js/helpers/native/wrapperTest/nativeTestImage');
var nativeExtCall = require('syracuse-tablet/html/js/helpers/native/nativeExtCall');

var _currentDeferred;

// Global object to be called by native code on pencil events
if (!(window.smImageJS)) {
	window.smImageJS = {

		/*
		 *  data:
		 *  {
		 *  	Action: "ok" || "cancelled" || "error"
		 *  	ImageData: "data:image/png;base64,iVBORw0KGgoAAAANSUhE....",
		 *  	Message: "Error message in case action=error"
		 *  }
		 */
		getPictureFromCameraFinished: function(data) {
			log && log("Got image!");
			if (_currentDeferred) {
				var def = _currentDeferred;
				_currentDeferred = null;
				log && log("Got image: " + JSON.stringify(data || {}));
				def.resolve({
					action: data.Action,
					imageData: data.ImageData
				});
			}
		},
		/*
		 *  	Action: "ok" || "error" || "unchanged"
		 *  	ImageData: "data:image/png;base64,iVBORw0KGgoAAAANSUhE....",
		 *  	Width: width
		 *  	Height: height
		 */
		getScaledImageFinished: function(data) {
			log && log("Got scaled image!");
			if (_currentDeferred) {
				var def = _currentDeferred;
				_currentDeferred = null;
				log && log("Got scaled image: " + JSON.stringify(data || {}));
				def.resolve({
					action: data.Action,
					imageData: data.ImageData,
					width: data.Width,
					height: data.Height
				});
			}
		}
	};
}
/**
 *
 */
exports.supports = function(capability) {
	if (capability === "camera") {
		return nativeExtCall.isSupported("smImage", "getPictureFromCameraSupported");
	}
	if (capability === "imageScale") {
		return nativeExtCall.isSupported("smImage", "getScaledImageSupported");
	}
	return false;
};

/**
 * Options are available as parameter but it's not used yet!
 * options:
 * {
 * }
 */
exports.getPictureFromCamera = function(options) {
	// No options yet!
	var param = {};
	_currentDeferred = $.Deferred();
	nativeExtCall.fireMethod("smImage", "getPictureFromCamera", param);
	return _currentDeferred.promise();
};

/**
 * options: {
 *   imageData: Data of image to scale: data:image/png;base64,iVBORw0KGgoAAAANSUh
 *   width: Desired width
 *   height: Desired height,
 *
 *   mode: "LIMIT_MAX_WIDTH_AND_HEIGHT"
 * }
 *
 *  Modes:
 *  LIMIT_MAX_WIDTH_AND_HEIGHT:
 *  Ensures that width and height do not exceed the values given in the options
 *  If one of the two is bigger that desired, both are scaled down to preserver the aspect ratio
 */
exports.getScaledImage = function(options) {
	var data = {
		ImageData: options.imageData,
		Width: options.width,
		Height: options.height,
		Mode: options.mode
	};
	_currentDeferred = $.Deferred();
	nativeExtCall.fireMethod("smImage", "getScaledImage", data);
	return _currentDeferred.promise();
};
/**
 * Returns the test js wrapper
 */
exports.init = function(testMode) {
	if (window && window.smImage) return;
	if (!testMode) return;
	log && log("Creating native image test wrapper");
	window.smImage = nativeTestImage.create();
	return window.smImage;
};
