// syracuse-tablet/html/js/helpers/native/nativePenInput.js
"use strict";

var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("nativePenInput");
var nativeTestPenInput = require('syracuse-tablet/html/js/helpers/native/test/nativeTestPenInput');
var nativeExtCall = require('syracuse-tablet/html/js/helpers/native/nativeExtCall');

var _currentInputDeferred;

// Global object to be called by native code on pencil events
if (!(window.smPenInputJS)) {
	window.smPenInputJS = {

		/*
		 *  This method will be invoked from wrapper once the pen input has finished or failed
		 *  data:
		 *  {
		 *  	Action: "ok" | "cancelled",
		 *  	Matches: ["match 1", "match 2"]
		 *  }
		 */
		penInputTextFinished: function(data) {
			if (_currentInputDeferred) {
				var def = _currentInputDeferred;
				_currentInputDeferred = null;
				log && log("Got pen input: " + JSON.stringify(data || {}));
				def.resolve({
					action: data.Action,
					matches: data.Matches
				});
			}
		},
		/*
		 *  This method will be invoked from wrapper once the user
		 *  finished drawing lines on the image he wants to draw annotations on
		 *  data:
		 *  {
		 *  	Action: "ok" | "cancelled",
		 *  	ImageData: "base64string with new image data" (data:image/jpeg;base64,AFHXSJKAHSFKLJH)
		 *  }
		 */
		penAnnotateImageFinished: function(data) {
			if (_currentInputDeferred) {
				var def = _currentInputDeferred;
				_currentInputDeferred = null;
				def.resolve({
					action: data.Action,
					imageData: data.ImageData
				});
			}
		},

		/*
		 *  This method will be invoked from wrapper once the user
		 *  finished to draw his signature
		 *  data:
		 *  {
		 *  	Action: "ok" | "cancelled",
		 *  	ImageData: "base64string with new image data" (data:image/jpeg;base64,AFHXSJKAHSFKLJH)
		 *  }
		 */
		penSignatureImageFinished: function(data) {
			if (_currentInputDeferred) {
				var def = _currentInputDeferred;
				_currentInputDeferred = null;
				def.resolve({
					action: data.Action,
					imageData: data.ImageData
				});
			}
		}
	};
}

/**
 *
 */
exports.supports = function(capability) {
	log && log("Pen check support: " + capability);
	if (capability === "penTextInput") {
		return nativeExtCall.isSupported("smPenInput", "GetPenTextInputSupported");
	}
	if (capability === "penAnnotatedImage") {
		return nativeExtCall.isSupported("smPenInput", "GetAnnotatedImageSupported");
	}
	if (capability === "penSignatureImage") {
		return nativeExtCall.isSupported("smPenInput", "GetSignatureImageSupported");
	}
	return false;
};

/**
 * Trigger pen input modal in native shell
 * options:
 * {
 *      title: Title to display above input area,
 *      value: Current field value // Must be of type string!
 * }
 */
exports.getPenTextInput = function(options) {
	if (options && options.value != null) {
		options.value = "" + options.value;
	}
	var data = JSON.stringify({
		Title: options.title,
		Value: options.value
	});
	_currentInputDeferred = $.Deferred();
	nativeExtCall.fireMethod("smPenInput", "GetPenTextInput", data);
	return _currentInputDeferred.promise();
};

/**
 * Trigger pen input on an image to allow highlighting with pencil
 * options:
 * {
 *      title: Title of image property
 *      imageData: base64 string containing image data (data:image/jpeg;base64,AFHXSJKAHSFKLJH)
 * }
 */
exports.getAnnotatedImage = function(options) {
	var data = JSON.stringify({
		Title: options.title,
		ImageData: options.imageData
	});
	_currentInputDeferred = $.Deferred();
	nativeExtCall.fireMethod("smPenInput", "GetAnnotatedImage", data);
	return _currentInputDeferred.promise();

};

/**
 * Trigger pen input to allow user to input his signature
 * options:
 * {
 *      title: Title of page
 *      watermark: Text that will be embedded in the image (e.g. a timestamp)
 *		maxWidth: 640 - Max width of signature image in pixels
 *		maxHeight: 480 - Max height of signature image in pixels
 *      watermarkSize: 16 - Height of watermark text. If text is too long or to high to fit, this size may be shrinked
 * }
 */
exports.getSignatureImage = function(options) {
	var data = JSON.stringify({
		Title: options.title,
		Watermark: (options.watermark != null ? ("" + options.watermark) : ""),
		MaxWidth: (options.maxWidth != null ? (+options.maxWidth) : 640),
		MaxHeight: (options.maxHeight != null ? (+options.maxHeight) : 480),
		WatermarkSize: (options.watermarkSize != null ? (+options.watermarkSize) : 16)
	});
	_currentInputDeferred = $.Deferred();
	nativeExtCall.fireMethod("smPenInput", "GetSignatureImage", data);
	return _currentInputDeferred.promise();

};

exports.init = function(testMode) {
	if (window && window.smPenInput) return;
	if (!testMode) return;
	log && log("Creating native Pen Input test wrapper");
	window.smPenInput = nativeTestPenInput.create();
};
