// syracuse-tablet/html/js/helpers/native/nativeNotification.js
"use strict";

var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("nativeLiveTile", true);

var _currentDeferred;
//Global object to be called by native code on pencil events
if (!(window.smNotificationsJS)) {
	window.smNotificationJS = {

		/*
		 *  This method will be invoked from wrapper when there was a live tile action executed.
		 *  data {
		 *  	action :"add" //"remove" "get"
		 *  	content:{ (null in case of get if not tile found)
		 *  			id:"id passed",
		 *  			title:"my customers",
		 *  			...
		 * 		}
		 *  	error:{ (if error occurs)
		 *  		no:"132001"
		 *  		text:"Not found", "unexpected error" ...
		 *  	}
		 *  }
		 * On error return is:
		 * {
		 *   error: {
		 *     text: "<error removing tile, for reason xxx>"
		 *   }
		 * }
		 */
		addNotificationFinished: function(data) {
			if (_currentDeferred) {
				var def = _currentDeferred;
				_currentDeferred = null;
				log && log("Command Notification add finished: " + JSON.stringify(data || {}));
				def.resolve(data);
			}
		},

		/**
		 * data: {
		 *    exists: true  | false
		 * }
		 *
		 * Return: true - tile exists
		 */
		getNotificationFinished: function(data) {
			if (_currentDeferred) {
				var def = _currentDeferred;
				_currentDeferred = null;
				log && log("Command Notification get finished: " + JSON.stringify(data || {}));
				def.resolve(data && data.exists);
			}
		},

		/**
		 * On error return is:
		 * {
		 *   error: {
		 *     text: "<error removing tile, for reason xxx>"
		 *   }
		 * }
		 */
		removeNotificationFinished: function(data) {
			if (_currentDeferred) {
				var def = _currentDeferred;
				_currentDeferred = null;
				log && log("Command Notification remove finished: " + JSON.stringify(data || {}));
				def.resolve();
			}
		}
	};
}

/**
 *
 */
exports.supportsNotification = function() {
	var supported = (window && window.smNotification && window.smNotification.supportsNotification && window.smNotification.supportsNotification()) || false;
	log && log("Notification : " + supported);
	return supported;
};

/**
	options:{

	}
	
	Return see addLiveTileFinished
 */
exports.addNotification = function(options) {
	if (window && window.smNotification && window.smNotification.addNotification) {
		_currentDeferred = $.Deferred();
		log && log("addNotification:" + JSON.stringify(options));
		var data = JSON.stringify(_getTileDataStruct(options));
		window.smLiveTile.addLiveTile(data);
		return _currentDeferred.promise();
	} else {
		return $.smResolve();
	}
};

/**
 *
 * See getLiveTileFinished, will return true or false depending if the tile exists or not
 */
exports.getNotification = function(id) {
	if (window && window.smLiveTile && window.smLiveTile.getLiveTile) {
		_currentDeferred = $.Deferred();
		var data = JSON.stringify(_getDataStruct({
			id: id
		}));
		window.smLiveTile.getNotification(data);
		return _currentDeferred.promise();
	} else {
		return $.smResolve();
	}
};

/**
 *	Return see removeLiveTileFinished
 */
exports.removeNotification = function(id) {
	if (window && window.smLiveTile && window.smLiveTile.removeLiveTile) {
		_currentDeferred = $.Deferred();
		var data = JSON.stringify(_getTileDataStruct({
			id: id
		}));
		window.smLiveTile.removeLiveTile(data);
		return _currentDeferred.promise();
	} else {
		return $.smResolve();
	}
};

/*
 * Pack everything into a structure understood by native code
 */
function _getDataStruct(options) {
	var param = {
		Id: options.id,
		Title: options.Title,
		Description: options.Description,
		SdataUrl: options.sDataUrl,
		DeepLink: options.DeepLink
	};
	return param;
}
