// syracuse-tablet/html/js/helpers/native/nativeImage.js
"use strict";

var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("nativeShare");
var utils = require('syracuse-tablet/html/js/helpers/utils');
var jsutils = require('syracuse-tablet/html/js/helpers/jsutils');
var nativeTestImage = require('syracuse-tablet/html/js/helpers/native/test/nativeTestImage');
var nativeHelpers = require('syracuse-tablet/html/js/helpers/native/nativeHelpers');

var _currentShareDeferred;

// Global object to be called by native code on pencil events
if (!(window.smImageJS)) {
	window.smImageJS = {

		/*
		 *  data:
		 *  {
		 *  	Action: "ok" || "cancelled",
		 *  	ImageData: "data:image/png;base64,iVBORw0KGgoAAAANSUhE...."
		 *  }
		 */
		getPictureFromCameraFinished: function(data) {
			if (_currentShareDeferred) {
				var def = _currentShareDeferred;
				_currentShareDeferred = null;
				log && log("Got image: " + JSON.stringify(data || {}));
				def.resolve({
					action: data.Action,
					imageData: data.ImageData
				});
			}
		}
	};
}
/**
 *
 */
exports.supports = function(capability) {
	if (capability === "camera") {
		var supported = (window && window.smImage && window.smImage.supportsPictureFromCamera && window.smImage.supportsPictureFromCamera()) || false;
		log && log("Share support: " + supported);
		return supported;
	}
	return false;
};

/**
 * options:
 * {
 * }
 */
exports.getPictureFromCamera = function(options) {
	// No options yet!
	var param = {};
	var data = JSON.stringify(param);
	log && log("getPictureFromCamera: ", data);
	// Be carefull, JS name is lowercase (e.g. setPhraseList) while C# name must be uppercase (e.g. SetPhraseList)
	if (window && window.smShare && window.smImage.getPictureFromCamera) {
		_currentShareDeferred = $.Deferred();
		window.smImage.getPictureFromCamera(data);
		return _currentShareDeferred.promise();
	} else {
		return $.smResolve(null);
	}
};

exports.init = function(testMode) {
	if (window && window.smImage) return;
	if (!testMode) return;
	log && log("Creating native image test wrapper");
	window.smImage = nativeTestImage.create();
};
