// syracuse-tablet/html/js/helpers/native/nativeTestWrapper.js
"use strict";

var localStorage = require('syracuse-tablet/html/js/storage/localStorage');

var _TestNativeWrapper = function() {
	/**
	 *
	 */
	this.onWebappLoaded = function() {};
	/**
	 *
	 */
	this.onWebappConfigServer = function() {};
	/**
	 *
	 */
	this.setReconnectContext = function(context) {
		try {
			localStorage.setItem("$reconnectContext", context ? JSON.stringify(context, null, 2) : null);
		} catch (e) {
			console.log(e);
		}
	};
	/**
	 *
	 */
	this.getReconnectContext = function() {
		try {
			var context = localStorage.getItem("$reconnectContext");
			return context == null ? context : JSON.parse(context);
		} catch (e) {
			console.log(e);
			return null;
		}
	};
	/**
	 *
	 */
	this.getReconnectToken = function() {
		try {
			var token = localStorage.getItem("$reconnectToken");
			return token == null ? token : JSON.parse(token);
		} catch (e) {
			console.log(e);
			return null;
		}
	};
	/**
	 *
	 */
	this.setReconnectToken = function(token) {
		try {
			localStorage.setItem("$reconnectToken", token ? JSON.stringify(token, null, 2) : null);
		} catch (e) {
			console.log(e);
		}
	};
	/**
	 *
	 */
	this.testClosingApplication = function(context) {
		if (!exports.isTestNativeWrapper()) return;
		try {
			localStorage.setItem("$reconnectHistory", JSON.stringify(context, null, 2));
		} catch (e) {}
	};
	/**
	 *
	 */
	this.getHistoryContext = function() {
		try {
			return localStorage.getItem("$reconnectHistory");
		} catch (e) {
			console.log(e);
			return null;
		}
	};
};
exports.create = function() {
	return new _TestNativeWrapper();
};
