// syracuse-tablet/html/js/helpers/native/nativeLiveTile.js
"use strict";

var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("nativeLiveTile", true);
var nativeTestLiveTile = require('syracuse-tablet/html/js/helpers/native/test/nativeTestLiveTile');


/*
 * Pack everything into a structure understood by native code
 */
function _getTileDataStruct(options) {
	var param = {
		Id: options.id,
		Title: options.Title,
		Description: options.Description,
		SdataUrl: options.sDataUrl,
		DeepLink: options.DeepLink
	};
	return param;
}

var _currentDeferred;
//Global object to be called by native code on pencil events
if (!(window.smLiveTileJS)) {
	window.smLiveTileJS = {

		/*
		 *  This method will be invoked from wrapper when there was a live tile action executed.
		 *  data {
		 *  	action :"add" //"remove" "get"
		 *  	content:{ (null in case of get if not tile found)
		 *  			id:"id passed",
		 *  			title:"my customers",
		 *  			...
		 * 		}
		 *  	error:{ (if error occurs)
		 *  		no:"132001"
		 *  		text:"Not found", "unexpected error" ...
		 *  	}
		 *  }
		 * On error return is:
		 * {
		 *   error: {
		 *     text: "<error removing tile, for reason xxx>"
		 *   }
		 * }
		 */
		addLiveTileFinished: function(data) {
			if (_currentDeferred) {
				var def = _currentDeferred;
				_currentDeferred = null;
				log && log("Command live tile add finished: " + JSON.stringify(data || {}));
				def.resolve(data);
			}
		},

		/**
		 * data: {
		 *    exists: true  | false
		 * }
		 *
		 * Return: true - tile exists
		 */
		getLiveTileFinished: function(data) {
			if (_currentDeferred) {
				var def = _currentDeferred;
				_currentDeferred = null;
				log && log("Command live tile get finished: " + JSON.stringify(data || {}));
				def.resolve(data && data.exists);
			}
		},

		/**
		 * On error return is:
		 * {
		 *   error: {
		 *     text: "<error removing tile, for reason xxx>"
		 *   }
		 * }
		 */
		removeLiveTileFinished: function(data) {
			if (_currentDeferred) {
				var def = _currentDeferred;
				_currentDeferred = null;
				log && log("Command live tile remove finished: " + JSON.stringify(data || {}));
				def.resolve();
			}
		}
	};
}

/**
 *
 */
exports.supports = function(capability, testMode) {
	if (capability === "liveTile") {
		var supported = (window && window.smLiveTile && window.smLiveTile.supportsLiveTile && window.smLiveTile.supportsLiveTile()) || false;
		log && log("Live tile : " + supported);
		return supported;
	}
	return false;
};

/**
	options:{
		"name":"x3.erp.GX3APP.BPCUSTOMERM.$query",
		"title":"Clients",
		"id":"x3.erp.GX3APP.BPCUSTOMERM.$query_http://pc101458:8124/mobile1/x3/erp/GX3APP/BPCUSTOMER?representation=BPCUSTOMERM.$query",
		"icon":"star",
		"sDataUrl":"http://pc101458:8124/mobile1/x3/erp/GX3APP/BPCUSTOMER?representation=BPCUSTOMERM.$query",
		"representation":"BPCUSTOMERM",
		"facet":"$query",
		"action":"$query",
		"entity":"BPCUSTOMER",
		"endpoint":"x3.erp.GX3APP",
		"applicationName":"sagesalescustomers"
	}
	
	Return see addLiveTileFinished
 */
exports.addLiveTile = function(options) {
	if (window && window.smLiveTile && window.smLiveTile.addLiveTile) {
		_currentDeferred = $.Deferred();
		log && log("addLiveTile:" + JSON.stringify(options));
		var data = JSON.stringify(_getTileDataStruct(options));
		window.smLiveTile.addLiveTile(data);
		return _currentDeferred.promise();
	} else {
		return $.smResolve();
	}
};

/**
 *
 * See getLiveTileFinished, will return true or false depending if the tile exists or not
 */
exports.getLiveTile = function(id) {
	if (window && window.smLiveTile && window.smLiveTile.getLiveTile) {
		_currentDeferred = $.Deferred();
		var data = JSON.stringify(_getTileDataStruct({
			id: id
		}));
		window.smLiveTile.getLiveTile(data);
		return _currentDeferred.promise();
	} else {
		return $.smResolve();
	}
};

/**
 *	Return see removeLiveTileFinished
 */
exports.removeLiveTile = function(id) {
	if (window && window.smLiveTile && window.smLiveTile.removeLiveTile) {
		_currentDeferred = $.Deferred();
		var data = JSON.stringify(_getTileDataStruct({
			id: id
		}));
		window.smLiveTile.removeLiveTile(data);
		return _currentDeferred.promise();
	} else {
		return $.smResolve();
	}
};

exports.init = function(testMode) {
	if (window && window.smLiveTile) return;
	if (!testMode) return;
	log && log("Creating native live tile test wrapper");
	window.smLiveTile = nativeTestLiveTile.create();
};
