// @sage/syracuse-ui/lib/search/searcher.js
"use strict";
var helpers = require('@sage/syracuse-core').helpers;
var Field = require('@sage/syracuse-ui/lib/field/field').Field;

function SuggestBox(field, $source, slot) {
	this.$source = $source;
	this.field = field;
	this.isSuggest = true;
	this.$totalResults = 0;

	this.slot = syra_dom.div("s-suggest-box", slot);

	this.header = syra_dom.text("s-suggest-box-title", this.$title = syra_local["search_" + ($source == "data" ? "documents" : $source)], this.slot);

	this.body = syra_dom.ul("s-suggest-box-items", this.slot);

	this.fetch = function(value) {
		var self = this;
		syra_dom.empty(self.body);
		self.rows = [];
		self._setEmptyMessage(syra_local.flWaitingData);
		self.field.fetchSearch({
			$source: self.$source,
			value: value,
			count: 5,
			options: {
				end: function(response) {
					if (!self.disposed) {
						if (response.ok) {
							self.fill(response.data);
						} else {
							self._setEmptyMessage(syra_local.search_no_res);
						}
					}
				}
			}
		});
	};

	this.fill = function(data) {
		syra_dom.empty(this.body);
		this.rows = [];
		this.$records = data.$resources || [];
		if (data.$totalResults != undefined) {
			this.$totalResults = data.$totalResults;
		}
		if (this.$totalResults) {
			this.header.textContent = this.$title + " (" + this.$totalResults + ")";
			var isData = this.$source == "data";
			for (var ii = 0, jj = this.$records.length; ii < jj; ii++) {
				var $record = this.$records[ii];
				var li = syra_dom.li("s-suggest-box-item");
				var row = syra_dom.a("s-suggest-box-item-link", li);
				this.rows.push(row);
				row.setAttribute("href", "#"); //need to improve cancel action otherwiese add # to url
				row.syraOnClick = "onSuggestClick";
				row.syraIndex = ii;
				row.syraBox = this.$source;
				row.syraItem = row.syrainout = this.field.id;
				row.appendChild(this.field.drawRecordText("s-suggest-box-item-title", $record.$searchItemTitle, $record));
				isData && row.appendChild(this.field.drawRecordText("s-suggest-box-item-type", $record.$resultTypeTitle, $record));
				this.body.appendChild(li);
			}
		} else {
			this.header.textContent = this.$title;
			this._setEmptyMessage(syra_local.search_no_res);
		}
		var count = (this.field.suggestBox.functions.$totalResults + this.field.suggestBox.data.$totalResults);
		syra_button.hide(this.field.suggestMore, count <= 0);
		syra_button.setText(this.field.suggestMore, count > 5 ? syra_local.searchAdvanceResults : syra_local.searchAdvanceDetails);
	};
	this._setEmptyMessage = function(text) {
		syra_dom.empty(this.body);
		syra_dom.text("s-suggest-box-empty", text, this.body);
	};
}


function SearchField() {}

helpers.defineClass(SearchField, Field, {
	getUrl: function($seach) {
		return syra_config.$searchUrl + ("&q=" + $seach) + (this.isFuzzyMode ? "&qt=fuzzy" : "") + ("&profile={$userprofile}");
	},
	drawRecordText: function(css, data, dataset) {
		var div = syra_dom.div(css);
		var text = syra_expression.parse(this.page, data, dataset);
		div.innerHTML = syra_core.escape(text.replace(/<em>/g, "<em class='" + css + "-em'>"));
		return div;
	},
	load: function() {
		var self = this;
		self.results = {};
		Field.prototype.load.call(self);

		self.clearBtn = syra_button.add({
			parent: self,
			slot: self._dataValue,
			text: syra_local.siteSearchClear,
			css: "s_sagearmonyeicon s-search-field-clear",
			iconOnly: true,
			click: function() {
				self.setValue("");
				self.applyEscape();
			}
		});
		self.clearBtn.link.textContent = "close"; // hack for sagearmoyeicon
		setTimeout(function() {
			self.focus();
		}, 100);
	},
	showPicker: function(onOut) {
		var show = this.getValue() != "";
		if (show) {
			this.clearBtn.link.style.opacity = onOut ? 0.4 : 1;
		}
		syra_button.hide(this.clearBtn, !show);
	},
	onKeyUp: function() {
		// this.showPicker();
	},
	onInputFocus: function(input, focusin) {
		// this.showPicker(!focusin);
	},
	onRunClick: function(results) {
		this.results = results || {}; //ensure reset of results
		var value = this.getValue();
		if (value !== "") {
			syra_dataset.setFieldValue(this, value);
			this._runSearch(this.getUrl(value));
		}
	},
	onSuggestClick: function(event, target) {
		var box = this.suggestBox && this.suggestBox[target.syraBox];
		this.executeRecord(box && box.$records[target.syraIndex]);
	},
	executeRecord: function($record) {
		$record && $record.$url && this._runSearch($record.$url, $record);
	},
	_runSearch: function($url, $record) {
		clearTimeout(this._searchTimer);
		syra_router.executeMenu({
			$url: syra_expression.parse(this.page, $url, $record)
		}, this.page);
		this.closeSuggest();
	},

	closeSuggest: function() {
		this.suggestPopup && this.suggestPopup.close();
	},
	_openSuggest: function() {
		var self = this;
		var value = self.getValue();
		if (value !== "") {
			self.page.dataset.$search = value;
			if (self.searchValue !== value) {
				self.focus();
				self.searchValue = value;
				if (!self.suggestPopup) {
					var slot = syra_dom.div("s-suggest");
					self.suggestBox = {
						"functions": new SuggestBox(self, "functions", slot),
						"data": new SuggestBox(self, "data", slot)
					};

					self.suggestMore = syra_button.add({
						parent: self,
						slot: slot,
						text: syra_local.searchAdvanceDetails,
						css: "s-suggest-more",
						click: function() {
							this.parent.onRunClick();
						}
					});
					syra_site.siteApp.appendChild(slot);

					self.suggestPopup = syra_popup.open({
						isAutoClose: false,
						content: self,
						picker: self.layoutSlot,
						slot: slot,
						my: "right top",
						at: "right bottom",
						of: self.domItem,
						close: function() {
							self.searchValue = null;
							syra_dom.empty(slot);
							if (self.suggestBox) {
								syra_site.disposeObject(self.suggestBox.functions);
								syra_site.disposeObject(self.suggestBox.data);
							}
							slot = self.suggestBox = self.suggestMore = self.suggestPopup = self.suggestSelected = null;
						}
					});
				}
				self.suggestBox.functions.fetch(value);
				self.suggestBox.data.fetch(value);
			}
		} else {
			self.closeSuggest();
		}
	},
	onItemInOut: function(on, event, target) {
		Field.prototype.onItemInOut.call(this, on, event, target);
		if (this.suggestSelected) {
			syra_dom.toggleClass(this.suggestSelected, "s-enter", false);
			this.suggestSelected = null;
		}
		syra_dom.toggleClass(target, "s-enter", on);
	},
	applyEscape: function() {
		clearTimeout(this._searchTimer);
		this.suggestPopup && this.closeSuggest();
		this.onHidden(this.getValue());
	},
	applyShortCut: function(shortcuts) {
		var self = this;
		var isCtrlPress = shortcuts.ctrl;
		clearTimeout(self._searchTimer);
		if (!shortcuts.enter) {
			self._searchTimer = setTimeout(function() {
				self._openSuggest();
			}, 1000);
			if (self.suggestBox && self.suggestPopup && (shortcuts.up || shortcuts.down)) {
				var rows = self.suggestBox.functions.rows.concat(self.suggestBox.data.rows).concat(self.suggestMore.link);
				if (rows.length > 1) {
					var next = 0;
					if (self.suggestSelected) {
						next = rows.indexOf(self.suggestSelected);
						if (next >= 0) {
							next += (shortcuts.down ? 1 : (-1));
						}
						syra_dom.toggleClass(self.suggestSelected, "s-enter", false);
						delete self.suggestSelected;
					}
					if (shortcuts.down) {
						self.suggestSelected = rows[(next >= rows.length) ? 0 : next];
					} else {
						self.suggestSelected = rows[(next < 0) ? (rows.length - 1) : next];
					}
					syra_dom.toggleClass(self.suggestSelected, "s-enter", true);
				}
			}
		} else {
			if (self.suggestPopup && self.suggestSelected && self.suggestSelected.parentNode) {
				self.suggestSelected.click();
				return true;
			}
			self.closeSuggest();
			self.results = {};
			var value = self.getValue();
			if (value != "") {
				["functions", "data"].forEach(function($source) {
					self.fetchSearch({
						$source: $source,
						value: value,
						options: {
							end: function(response) {
								if (!self.disposed) {
									if (response.ok) {
										self.results[$source] = response.data;
										if (self.results.functions && self.results.data) {
											var $records = self.results.functions.$resources;
											if ($records && $records.length == 1) {
												var $record = $records[0];
												var code = $record.fusionFunction && $record.fusionFunction.replace("<em>", "").replace("</em>", "");
												var run = (code && code.toUpperCase() == value.toUpperCase());
												if (run && (self.results.data.$resources && self.results.data.$resources.length > 0) && !isCtrlPress) {
													run = false;
												}
												if (run) {
													self.results = {}; //reset
													self.executeRecord($record);
													return;
												}
											}
											self.onRunClick(self.results);
										}
									} else {
										syra_alert.show(response.error.diagnoses, self.page);
									}
								}
							}
						}
					});
				});
			}
		}
		return true;
	},
	fetchSearch: function(context) {
		var self = this;
		context.options = context.options || {};
		context.options.page = self.page;
		var $url = self.getUrl(context.value) + "&source=" + context.$source;
		if (context.count) {
			$url += "&count=" + context.count;
		}
		syra_url.formatMenuUrl(self.page, {
			$url: $url
		}, {}, function($fullUrl, isCanceled) {
			context.options.url = $fullUrl;
			syra_ajax.get(context.options);
		});
	},
	dispose: function() {
		this.closeSuggest();
		Field.prototype.dispose.call(this);
	}

});

exports.load = function(slot, value, onHidden) {
	syra_site.$prototype.$properties.$search = {
		$type: "application/x-string",
		$title: syra_local.searchFieldTitle
	};
	var field = syra_site.searcher = syra_site.addItem(slot, {
		$bind: "$search",
		$isEditMode: true,
		$class: SearchField,
		$skin: "s-search-field",
		$inplace: true,
		$isTitlePlaceHolder: true
	});
	field.onHidden = onHidden;
	return field;
};
