// syracuse-tablet/html/js/wc/pages/pageBase.js
"use strict";

var utils = require('syracuse-tablet/html/js/wc/helpers/utils');
var Controller = require("syracuse-tablet/html/js/wc/controllers/controller").Controller;
var AttachedControlsMgr = require('syracuse-tablet/html/js/wc/pages/attachedControlsMgr').AttachedControlsMgr;

var siteLayout = require('syracuse-tablet/html/js/wc/ui/siteLayout');
var uiRect = require('syracuse-tablet/html/js/wc/ui/rect');
var scroller = require('syracuse-tablet/html/js/wc/helpers/scroller/scrollerHelpers');

var _id = 1;
var _templates = {};
var _html = {
	fullpage: '\
		<section class="s-m-page s-m-full {{device}}" id="{{pageId}}" style="display: none" data-page-name="{{pageName}}"> \
			<header></header> \
			<section class="s-m-breadcrumbs" style="display:none"></section> \
			<section class="s-m-topToolbar"  style="display:none"></section> \
			<section class="s-m-main-content" data-s-m-content=""></section> \
			<footer style="display:none"></footer> \
		</section> \
	',
	vignette: '<section class= "s-m-page s-m-nested" id="{{pageId}}" style="display: none;" data-page-name="{{pageName}}"></section>',
	panelContainers: '\
		<div class="s-m-side-panel-dismiss" data-nevent data-naction="closePanel"></div> \
		<aside class="s-m-panel s-m-left"></aside> \
		<aside class="s-m-panel s-m-right"></aside>'

}
var _getHtml = function(name, ctx) {
	var tmpl = _templates[name];
	if (!tmpl) {
		tmpl = _templates[name] = Handlebars.compile(_html[name]);
	}
	return tmpl(ctx || {});
};

var _pages = {};

/**
 * options:
 * {
 *   isChild: true|false
 *   isVignette: true|false
 * }
 */
exports.Page = utils.defineClass(
	function PageBase(pageName, options) {
		this.pageName = pageName || "noname";
		this.id = _id++;
		_pages[this.id] = this;
		this.options = options || {};
		// set from outside, in here we want to load the page
		this.$$container = null;
		// root element of the page
		this.$$elmt = null;
		// element where the main page content is rendered to
		this.$$contentElmt = null;
		this.controller = null;
		// When a page is loaded inside a vignette, parentPage will point to the dashboard
		// containing the vignette
		this.parentPage = null;
		this.attachedControlsMgr = new AttachedControlsMgr(this);
		// Let page subclasses define their own controller if needed
		this._ensureController();
	},
	null, {
		/** 
		 * Override to create a specialized controller
		 */
		_ensureController: function() {
			this.controller = new Controller(null);
		},
		destroy: function() {
			this.unbuild();
			delete _pages[this.id];
		},

		/**
		 * Release resources that will be re-allocated on build
		 */
		unbuild: function() {
			if (this.attachedControlsMgr) {
				this.attachedControlsMgr.destroyControls();
			}

			if (this.$$elmt) {
				this.$$elmt.unbind();
				this.$$elmt.remove();
			}

			this.$$elmt = null;
			this.$$contentElmt = null; // Element in DOM which contains all content and eventually a scroller
			this.$$contentRoot = null; // Put content in here: Element in DOM that is eventually scrollable, could be the same as $$contentElmt if no scrolling is desired
		},
		// Return true whenever pushing this page onto the page stack
		// should clear the history and make this page the only one 
		// existing the the history (e.g. login, logout, home)
		isRootPage: function() {
			return false;
		},
		// For pages that must not be added to the history.
		// e.g. Login, logout
		isNoHistory: function() {
			return false;
		},
		set$$container: function($$container) {
			this.$$container = $$container;
		},

		/**
		 * Should be called by all subclasses
		 */
		createRootElement: function() {
			var tplName = "fullpage";
			if (this.options.isVignette) {
				tplName = "vignette";
			}

			var ctx = {
				pageId: this.id,
				pageName: this.pageName,
				device: siteLayout.getDeviceType(),
			};
			this.$$elmt = $(_getHtml(tplName, ctx));
			this.$$contentElmt = this.$$elmt.find('[data-s-m-content]'); // Is there a children flagged to be the content holding element
			if (this.$$contentElmt.length === 0) {
				this.$$contentElmt = this.$$elmt; // Use root element for page content if no other element is flagged to be used
			}
			this.$$contentRoot = this.$$contentElement;

			if (!this.options.isVignette) {
				this.$$elmt.append($(_getHtml("panelContainers")));
			}
			this.$$container.append(this.$$elmt);
			this._initScrolling();
		},

		hide: function() {
			if (!this.$$elmt) {
				return;
			}
			this.$$elmt.hide();
		},

		show: function() {
			if (!this.$$elmt) {
				return;
			}
			this.$$elmt.show();

			this._updateScrolling();
		},

		build: function() {
			var self = this;

			return $.smResolve()
				.then(function() {
					return self.createRootElement();
				})
				.then(function() {
					return self.initStructure && self.initStructure();
				})
				.then(function() {
					return self.render && self.render();
				})
				.then(function() {
					return self.afterRender && self.afterRender();
				})
				.then(function() {
					if (self.options.isVignette !== true) {
						self.computeLayout();
					}
				})
				.then(function() {
					self.onPageDataChanged();
				})
		},

		initStructure: function() {
			this.createAttachedControls();
		},

		computeLayout: function() {
			this.$$elmt.removeClass("landscape portrait");
			this.$$elmt.addClass(siteLayout.getPageOrientation());
			var layoutInfo = siteLayout.getLayoutInfo();
			if (this.options.isVignette !== true) {
				var mainLayout = this.attachedControlsMgr.computeMainLayout(layoutInfo);
				this.$$contentElmt.css({
					top: mainLayout.contentTop + "px",
					left: mainLayout.contentLeft + "px",
					width: mainLayout.contentWidth + "px",
					height: mainLayout.contentHeight + "px"
				});
			} else {
				/**
				 * For vignette we don't calculate the H/W in pixels.
				 * To do the calculatation we need to propagate the calculation from dashboard tile -> Control (not done and needed)
				 * H/W calculation is based on css style height auto for the wrapper et 100% for $$contentElmt
				 */
			}
			this._mainContentRect = uiRect.elmtRect(this.$$contentElmt, "computed");
			this._mainContentRect.moveLeft(parseInt(this.$$contentElmt.css("padding-left"), 10));
			layoutInfo.pageContentRect = this._mainContentRect;
			this.controller.computeLayout(layoutInfo);
			this._updateScrolling();
		},

		//
		// START: Scrolling stuff
		//

		/**
		 * Override for pages that do not want scrolling
		 */
		_initScrolling: function() {
			this.$$scrollWrapper = $('<div class="s-m-scroll-wrapper"/>').appendTo(this.$$contentElmt);
			this.$$scrollWrapper.css({
				height: "auto"
			});
			this.$$contentRoot = this.$$scrollWrapper;

			this._gestureMgr = scroller.newScroller(this.$$scrollWrapper, {
				isPageScroller: true // !self.isVignette
			});
		},
		_updateScrolling: function() {
			if (!this._gestureMgr) {
				return;
			}
			if (this.options.isVignette) {
				return;
			}
			var viewRect = this._mainContentRect;
			// Force $$scrollWrapper to full heigth of childrens
			// We scroll if children height is greater than childrenRect height
			if (viewRect && viewRect.height < this.$$scrollWrapper.height()) {
				var valMax = 0;
				this._gestureMgr.init(viewRect, "v", valMax, false);
			} else {
				this._gestureMgr.reset();
			}
		},

		/**
		 * Called by controls that have an impact on the page size after loading
		 * Currently, this are only vignettes
		 */
		onContentChanged: function() {
			this._updateScrolling();
		},

		//
		// END: Scrolling stuff
		//

		getPageTitle: function() {
			return "";
		},

		createAttachedControls: function() {
			var self = this;
			var names = self._getDefaultAttachedControls();
			names.forEach(function(name) {
				var info = self._getAttachedControlInfo(name);
				if (info) {
					self.attachedControlsMgr.createControl(info);
				}
			});
		},

		/**
		 * Get logical name of controls that are usually attached to a page
		 * 
		 * Can be overriden per page if necessarry
		 * @returns
		 */
		_getDefaultAttachedControls: function() {
			if (this.options.isVignette) {
				return [];
			}
			var device = siteLayout.getDeviceType();
			if (device === "smartphone") {
				return ["header"];
			} else {
				return ["header", "breadcrumbs", /*  "topToolbar",*/ "footer"];
			}
		},

		/**
		 * Get detail information on how to create a logical control
		 * E.g. device dependent header
		 * 
		 * Can be overriden per page if necessarry
		 * 
		 * @param name
		 * @returns
		 */
		_getAttachedControlInfo: function(name) {
			var device = siteLayout.getDeviceType();
			var defaults = {
				smartphone: {
					header: {
						topIndex: 0,
						selector: "header",
						$type: "application/x-panel-header-sdata-page-smartphone"
					},
					actionsPanel: {
						selector: "aside.s-m-panel.s-m-right",
						$type: "application/x-panel-actions"
					},
					globalPanel: {
						selector: "aside.s-m-panel.s-m-left",
						$type: "application/x-panel-global-smartphone"
					},
					filterSortPanel: {
						selector: "aside.s-m-panel.s-m-right",
						vignetteSelectorUseParent: true,
						$type: "application/x-panel-filtersort"
					}
				},
				tablet: {
					header: {
						topIndex: 0,
						selector: "header",
						$type: "application/x-panel-header-sdata-page-tablet"
					},
					breadcrumbs: {
						topIndex: 1,
						selector: ".s-m-breadcrumbs",
						$type: "application/x-panel-breadcrumbs-tablet"
					},
					topToolbar: {
						topIndex: 2,
						selector: ".s-m-topToolbar",
						$type: "application/x-panel-???" // TODO:
					},
					footer: {
						bottomIndex: 0,
						selector: "footer",
						$type: "application/x-panel-footer-sdata-page"
					},
					actionsPanel: {
						selector: "aside.s-m-panel.s-m-right",
						$type: "application/x-panel-actions"
					},
					globalPanel: {
						selector: "aside.s-m-panel.s-m-right",
						$type: "application/x-panel-global-tablet"
					},
					filterSortPanel: {
						selector: "aside.s-m-panel.s-m-right",
						vignetteSelectorUseParent: true,
						$type: "application/x-panel-filtersort"
					}
				}
			};

			var info = defaults[device][name];
			info.name = name;
			return info;
		},
		openPanel: function(name, options) {
			var panel = this.attachedControlsMgr.findByName(name);
			if (!panel) {
				var info = this._getAttachedControlInfo(name);
				if (info) {
					panel = this.attachedControlsMgr.createControl(info, options);
				}
			}
			if (!panel) {
				return;
			}
			return panel
		},
		_actOpenGlobalPanel: function() {
			var panel = this.openPanel("globalPanel");
			if (panel) panel.show();
		},
		/**
		 * Should do:
		 * - Destroy and re-create page specific controls like header/footer depending on the page size/device
		 *   (also this is mostly a development use case)
		 * - Close opened panels and popups 
		 */
		onMainPageResize: function() {
			this.computeLayout();
		},
		/**
		 * Called by navigator whenever the page will be closed because of a back action
		 * The back action will be done as soon as the given deferred is resolved with "true"
		 * Resolving with "false" will cancel the back action
		 * Rejecting will show the error given to reject
		 */
		onBackAction: function(def) {
			// Default is to accept back
			def.resolve(true);
		},

		onRequestUpdateScroller: function() {

		},
		/**
		 * Called when the data on a page has changed.
		 * First time called after page has rendered and the e.g. on paging
		 */
		onPageDataChanged: function() {},

		//
		// START: Authoring
		//
		allowAuthoring: function() {
			return false;
		},
		getAuthoringName: function() {

		},
		getPageName: function() {
			return this.pageName;
		},
		authUpdateLayout: function() {
				var self = this;
				self.unbuild();
				return self.build().then(function() {
					return self.show();
				});
			}
			//
			// END: Authoring
			//

	});

exports.findPage = function(id) {
	return _pages[id];
}
