// @sage/syracuse-react/src/page/page.tsx
"use strict";
import * as React from 'react';
import { Profile } from '../profile/profile';
import { Url } from '../route/url';
import './page.less';
import { OfficeController } from '../office/officeController';


export interface IPageProps {
    key: string,
    openedFrom?: Url,
    sdataUrl: Url,
    onMainPageLoaded?: (title: string, description?: string, syraPage?: any) => void
    profile: Profile,
    office?: OfficeController
}

export class PageFrame extends React.Component<{}, {}>{
    render() {
        return <article className="s_poc_page">{this.props.children}</article>;
    }
}


export class PageHeader extends React.Component<{
    title: string,
}, {}> {
    render() {
        return <header className='s_poc_page_header'>
            <div className='s_poc_page_title'>{this.props.title}</div>
        </header>;
    }
}

export class PageBody extends React.Component<{
    className?: string
}, {}> {
    render() {
        let css = "s_poc_page_body";
        if (this.props.className) {
            css += " " + this.props.className;
        }
        return <div className={css}>{this.props.children}</div>;
    }
}

export class WebPage extends React.Component<{
    title: string,
    bodyClassName?: string
}, {}>{
    render() {
        return <PageFrame>
            <PageHeader title={this.props.title} />
            <PageBody className={this.props.bodyClassName}>
                {this.props.children}
            </PageBody>
        </PageFrame>;
    }
}
