// syracuse-ui/lib/mobile/field/alphaNumField.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Field = require('syracuse-ui/lib/common/field/field').Field;
var commonField = require('syracuse-ui/lib/common/field/alphaNumField');
var fieldHelper = require('./fieldHelper');

function MobilePassword(){
}

exports.MobilePassword = helpers.defineClass(MobilePassword, Field, {
    getHtmlInput: function(){
        return "<input type='password'/>";
    },
    setDataValue: function(value, record){
        this.currentValue = value;
        if (this.$isEditMode) {
            this.$$input.val(this.currentValue);
        }
        else {
            this.$$dataValue.text((value || "").length > 0 ? "***" : "");
        }
    },
    loadBox: function(){
        fieldHelper.onLoadBox(Field, this);
    },
    setDescription: function(){
    },
    setHelp: function(){
    },
    _createValueContainer: function(){
        if (this.$isEditMode) {
            this.$$valueSlot = this.$$dataValue = this.$$item;
        }
        else {
            Field.prototype._createValueContainer.call(this);
        }
    }
});

function MobileNumeric(){
}

exports.MobileNumeric = helpers.defineClass(MobileNumeric, Field, {
    loadBox: function(){
        fieldHelper.onLoadBox(Field, this);
    },
    setDescription: function(){
    },
    setHelp: function(){
    },
    _createValueContainer: function(){
        if (this.$isEditMode) {
            this.$$valueSlot = this.$$dataValue = this.$$item;
        }
        else {
            Field.prototype._createValueContainer.call(this);
        }
    }
});

function MobileText(){
}

exports.MobileText = helpers.defineClass(MobileText, Field, {
    loadBox: function(){
        fieldHelper.onLoadBox(Field, this);
    },
    setDescription: function(){
    },
    setHelp: function(){
    },
    _createValueContainer: function(){
        if (this.$isEditMode) {
            this.$$valueSlot = this.$$dataValue = this.$$item;
        }
        else {
            Field.prototype._createValueContainer.call(this);
        }
    }
});

function MobileTextArea(){
}

exports.MobileTextArea = helpers.defineClass(MobileTextArea, MobileText, {
    getHtmlInput: function(){
        return "<textarea/>";
    }
});

function MobilePhone(){
}

exports.MobilePhone = helpers.defineClass(MobilePhone, commonField.PhoneField, {
    loadBox: function(){
        fieldHelper.onLoadBox(commonField.PhoneField, this);
    },
    setDescription: function(){
    },
    setHelp: function(){
    },
    _createValueContainer: function(){
        if (this.$isEditMode) {
            this.$$valueSlot = this.$$dataValue = this.$$item;
        }
        else {
            commonField.PhoneField.prototype._createValueContainer.call(this);
        }
    }
});

function MobileEmail(){
}

exports.MobileEmail = helpers.defineClass(MobileEmail, commonField.EmailField, {
    loadBox: function(){
        fieldHelper.onLoadBox(commonField.EmailField, this);
    },
    setDescription: function(){
    },
    setHelp: function(){
    },
    _createValueContainer: function(){
        if (this.$isEditMode) {
            this.$$valueSlot = this.$$dataValue = this.$$item;
        }
        else {
            commonField.EmailField.prototype._createValueContainer.call(this);
        }
    }
});
