// syracuse-ui/lib/mobile/field/alphaNumField.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var commonField = require('syracuse-ui/lib/common/field/alphaNumField');
var fieldHelper = require('./fieldHelper');

function MobilePassword(){
}

exports.MobilePassword = helpers.defineClass(MobilePassword, commonField.PasswordField, {
    loadBox: function(){
        fieldHelper.onLoadBox(commonField.NumericField, this);
    },
    setDescription: function(){
    },
    setHelp: function(){
    },
    _createValueContainer: function(){
        if (this.$isEditMode) {
            this.$$valueSlot = this.$$dataValue = this.$$item;
        }
        else {
            commonField.PasswordField.prototype._createValueContainer.call(this);
        }
    }
});

function MobileNumeric(){
}

exports.MobileNumeric = helpers.defineClass(MobileNumeric, commonField.NumericField, {
    loadBox: function(){
        fieldHelper.onLoadBox(commonField.NumericField, this);
    },
    setDescription: function(){
    },
    setHelp: function(){
    },
    _createValueContainer: function(){
        if (this.$isEditMode) {
            this.$$valueSlot = this.$$dataValue = this.$$item;
        }
        else {
            commonField.NumericField.prototype._createValueContainer.call(this);
        }
    }
});

function MobileText(){
}

exports.MobileText = helpers.defineClass(MobileText, commonField.TextField, {
    loadBox: function(){
        fieldHelper.onLoadBox(commonField.TextField, this);
    },
    setDescription: function(){
    },
    setHelp: function(){
    },
    _createValueContainer: function(){
        if (this.$isEditMode) {
            this.$$valueSlot = this.$$dataValue = this.$$item;
        }
        else {
            commonField.TextField.prototype._createValueContainer.call(this);
        }
    },
    setDataValue: function(value, record){
        this.currentValue = value;
        if (this.$isEditMode) {
            this.$$input.val(this.currentValue);
        }
        else {
            this.$$dataValue.empty();
            if (value && this.$field.$links && this.$field.$links.$details && !this.$item.$isMenusDisabled) {
                this.appendDetailLink(this.$$dataValue, value, this.$field.$links.$details);
            }
            else {
                this.$$dataValue.text(value);
            }
        }
    }
});

function MobileTextArea(){
}

exports.MobileTextArea = helpers.defineClass(MobileTextArea, MobileText, {
    getHtmlInput: function(){
        return "<textarea/>";
    },
    setDescription: function(){
    },
    setHelp: function(){
    }
});

function MobilePhone(){
}

exports.MobilePhone = helpers.defineClass(MobilePhone, commonField.PhoneField, {
    loadBox: function(){
        fieldHelper.onLoadBox(commonField.PhoneField, this);
    },
    setDescription: function(){
    },
    setHelp: function(){
    },
    _createValueContainer: function(){
        if (this.$isEditMode) {
            this.$$valueSlot = this.$$dataValue = this.$$item;
        }
        else {
            commonField.TextField.prototype._createValueContainer.call(this);
        }
    }
});

function MobileEmail(){
}

exports.MobileEmail = helpers.defineClass(MobileEmail, commonField.EmailField, {
    loadBox: function(){
        fieldHelper.onLoadBox(commonField.EmailField, this);
    },
    setDescription: function(){
    },
    setHelp: function(){
    },
    _createValueContainer: function(){
        if (this.$isEditMode) {
            this.$$valueSlot = this.$$dataValue = this.$$item;
        }
        else {
            commonField.TextField.prototype._createValueContainer.call(this);
        }
    }
});
