// syracuse-ui/lib/mobile/field/choiceField.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var ChoiceField = require('syracuse-ui/lib/common/field/choiceField').ChoiceField;
var fieldHelper = require('./fieldHelper');

function MobileComboChoice(){
}

exports.MobileComboChoice = helpers.defineClass(MobileComboChoice, ChoiceField, {
    loadBox: function(){
        fieldHelper.onLoadBox(ChoiceField, this, "selectmenu", function(field){
            if (field.$isEditMode) {
                field.$$input[0].setAttribute("data-overlay-theme", field.$theme.$overlay);
            }
        });
    },
    setDataValue: function(value){
        this.$selectedEnum = this.$listSelection = this._findEnum(value);
        var $title = this.$selectedEnum ? this.$selectedEnum.$title : "";
        if (this.$isEditMode) {
            if (this.currentValue && this.currentValue != value) {
                var $prevEnum = this._findEnum(this.currentValue);
                this.$$value.find("option[value='" + $prevEnum.$index + "']").attr("checked", false);
            }
            this.currentValue = value;
            if (this.$selectedEnum) {
                this.$$value.find("option[value='" + this.$selectedEnum.$index + "']").attr("selected", true);
            }
            if (document.site.isjqmPageLoaded()) {
                document.site.refreshJQMPlugin(this.$$item, "selectmenu");
            }
        }
        else {
            this.$$value.text($title);
        }
        
    },
    getInputValue: function(){
        return this.$selectedEnum ? this.$selectedEnum.$value : null;
    },
    render: function(){
        var self = this;
        self.$enum = self.$field.$value.$constraints.$enum;
        if (this.$isEditMode) {
            self.$$value = self.$$input = $("<select data-native-menu='false'/>").appendTo(self.$$item);
            if (self.inputId) {
                self.$$input.attr("id", self.inputId);
            }
            self.$enum.forEach(function($enum, index){
                var $$item = $("<option/>").attr("value", index).append($enum.$title);
                if (self.currentValue == $enum.$value) {
                    $$item.attr("selected", true);
                }
                self.$$input.append($$item);
                
            });
            self.$$item.children(".s-field-edit-value").remove();
        }
    },
	_createValueContainer:function(){
		if (this.$isEditMode) {
            this.$$value = this.$$item;
        }
        else {
            ChoiceField.prototype._createValueContainer.call(this);
        }
	}
});

function MobileRadioChoice(){
}

exports.MobileRadioChoice = helpers.defineClass(MobileRadioChoice, ChoiceField, {
    loadBox: function(){
        fieldHelper.onLoadBox(ChoiceField, this, "checkboxradio", function(field){
            if (field.$isEditMode) {
                if (document.site.isjqmPageLoaded()) {
                    field.$$fieldset.controlgroup({
                        excludeInvisible: false
                    });
                }
            }
        });
    },
    getInputValue: function(){
        var index = this.$$fieldset.find("input").filter(":checked").val();
        return (index == null) ? null : this.$enum[index].$value;
    },
    setDataValue: function(value){
        this.$selectedEnum = this._findEnum(value);
        var $title = this.$selectedEnum ? this.$selectedEnum.$title : "";
        if (this.$isEditMode) {
            if (this.currentValue && this.currentValue != value) {
                var $prevEnum = this._findEnum(this.currentValue);
                var $$prev = this.$$value.find("input[value='" + $prevEnum.$index + "']").attr("checked", false);
                if (this.$isjqmField) {
                    $$prev.checkboxradio("refresh");
                }
                if (document.site.isjqmPageLoaded()) {
                    document.site.refreshJQMPlugin($$prev, "checkboxradio");
                }
            }
            this.currentValue = value;
            if (this.$selectedEnum) {
                var $$new = this.$$value.find("input[value='" + this.$selectedEnum.$index + "']").attr("checked", true);
                if (document.site.isjqmPageLoaded()) {
                    document.site.refreshJQMPlugin($$new, "checkboxradio");
                }
            }
        }
        else {
            this.$$value.text($title);
        }
    },
    render: function(){
        var self = this;
        self.$enum = self.getDataType().$constraints.$enum;
        if (self.$isEditMode) {
            self.$$value = self.$$item;
            var $groupName = document.controller.generateUUID();
            self.$$fieldset = $("<fieldset data-role='controlgroup'" + (self.$item.halign ? 'data-type="horizontal"' : "") + "/>").appendTo(self.$$value);
            if (self.$$title) {
                self.$$title = $("<legend/>").appendTo(self.$$fieldset).append(self.$$title);
            }
            self.$enum.forEach(function($enum, index){
                var id = $groupName + index;
                var $$input = $("<input type='radio'/>").attr("id", id).attr("name", $groupName).val(index);
                var $$title = $("<label/>").attr("for", id).text($enum.$title);
                self.$$fieldset.append($$input).append($$title);
            });
            self.$$item.children(".s-field-edit-value").remove();
            
        }
    },
    applyDisableState: function(state){
        if (state.$isDisabled !== undefined) {
			this.$isDisabled = state.$isDisabled;
            this.$$item.attr("disabled", state.$isDisabled);
            this.$$value.find("input").attr("disabled", state.$isDisabled);
        }
    }
});
