// syracuse-tablet/html/js/demos/_entities/entCustomers.js
"use strict";
var utils = require('syracuse-tablet/html/js/helpers/utils');

var _names = [
	"Acme, inc.",
	"Widget Corp",
	"123 Warehousing",
	"Demo Company",
	"Smith and Co.",
	"Foo Bars",
	"ABC Telecom",
	"Fake Brothers",
	"QWERTY Logistics",
	"Demo, inc.",
	"Sample Company",
	"Sample, inc",
	"Acme Corp",
	"Allied Biscuit",
	"Ankh-Sto Associates",
	"Extensive Enterprise",
	"Galaxy Corp",
	"Globo-Chem",
	"Mr. Sparkle",
	"Globex Corporation",
	"LexCorp",
	"LuthorCorp",
	"North Central Positronics",
	"Omni Consimer Products",
	"Praxis Corporation",
	"Sombra Corporation",
	"Sto Plains Holdings",
	"Tessier-Ashpool",
	"Wayne Enterprises",
	"Wentworth Industries",
	"ZiffCorp",
	"Bluth Company",
	"Strickland Propane",
	"Thatherton Fuels",
	"Three Waters",
	"Water and Power",
	"Western Gas & Electric",
	"Mammoth Pictures",
	"Mooby Corp",
	"Gringotts",
	"Thrift Bank",
	"Flowers By Irene",
	"The Legitimate Businessmens Club",
	"Osato Chemicals",
	"Transworld Consortium",
	"Universal Export",
	"United Fried Chicken",
	"Virtucon",
	"Kumatsu Motors",
	"Keedsler Motors",
	"Powell Motors",
	"Industrial Automation",
	"Sirius Cybernetics Corporation",
	"U.S. Robotics and Mechanical Men",
	"Colonial Movers",
	"Corellian Engineering Corporation",
	"Incom Corporation",
	"General Products",
	"Leeding Engines Ltd.",
	"Blammo",
	"Input, Inc.",
	"Mainway Toys",
	"Videlectrix",
	"Zevo Toys",
	"Ajax",
	"Axis Chemical Co.",
	"Barrytron",
	"Carrys Candles",
	"Cogswell Cogs",
	"Spacely Sprockets",
	"General Forge and Foundry",
	"Duff Brewing Company",
	"Dunder Mifflin",
	"General Services Corporation",
	"Monarch Playing Card Co.",
	"Krustyco",
	"Initech",
	"Roboto Industries",
	"Primatech",
	"Sonky Rubber Goods",
	"St. Anky Beer",
	"Stay Puft Corporation",
	"Vandelay Industries",
	"Wernham Hogg",
	"Gadgetron",
	"Burleigh and Stronginthearm",
	"BLAND Corporation",
	"Nordyne Defense Dynamics",
	"Petrox Oil Company",
	"Roxxon",
	"McMahon and Tate",
	"Sixty Second Avenue",
	"Charles Townsend Agency",
	"Spade and Archer",
	"Megadodo Publications",
	"Rouster and Sideways",
	"C.H. Lavatory and Sons",
	"Globo Gym American Corp",
	"The New Firm"
];

var _customers = [];
for (var c = 10000; c < 10100; c++) {
	_customers.push({
		$uuid: "" + (c - 10000), // to match array index
		bpcnum: "" + c,
		name: _names[c - 10000],

		ordersValue: 1000000 + ((c - 10000) * 200000 % 500000),
		ordersMax: 1800000,
		quotesValue: 500000 + ((c - 10000) * 2000 % 50000),
		quotesMax: 1200000,
		invoicesValue: 10000 + ((c - 10000) * 2000 % 50000),
		invoicesMax: 120000,
		min: 1000,
		addr1: "Rue Fructidor",
		addr2: "75834 Paris Cedex 17"
	});
}

exports.entity = {
	$entityName: "$localCustomer",
	$value: "{name}",
	$properties: {
		bpcnum: {
			$type: "application/x-string",
			$title: "Customer"
		},
		name: {
			$type: "application/x-string",
			$title: "Name",
		},
		addr1: {
			$type: "application/x-string",
			$title: "Street",
		},
		addr2: {
			$type: "application/x-string",
			$title: "Address",
		},
		ordersValue: {
			$type: "application/x-decimal",
			$title: "Orders value"
		},
		min: {
			$type: "application/x-decimal",
			$title: "General min value"
		},
		quotesValue: {
			$type: "application/x-decimal",
			$title: "Quotes value"
		},
		invoicesValue: {
			$type: "application/x-decimal",
			$title: "Invoices value"
		},
		ordersMax: {
			$type: "application/x-decimal",
			$title: "Max"
		},
		quotesMax: {
			$type: "application/x-decimal",
			$title: "Max"
		},
		invoicesMax: {
			$type: "application/x-decimal",
			$title: "Max"
		}
	},
	$prototype: function(proto) {
		// Add some chart information in $query prototype
		var res = utils.getPropByPath(proto, "$properties.$resources.$item.$links") || {};
		// Link to use when user click on row detail link
		res["DASHBOARD"] = {
			$title: "Detail dashboard",
			$url: "{$baseUrl}/$mobileDashboards?dashboard=$localCustomerDashboard&parameters=bpcnum%3D%7Bbpcnum%7D"
		};
	},
	$articles: {
		$query: {
			"$layoutType": "stack",
			"$items": [{
				"$bind": "$resources",
				"$detailsLink": "DASHBOARD",
				"$display": "card",
				"$cardsPerLine": 4,
				"$article": {
					"$items": [{
						"$bind": "bpcnum"
					}, {
						"$bind": "name"
					}, {
						"$bind": "addr1"
					}, {
						"$bind": "addr2"
					}]
				}
			}]
		}
	},
	$services: {
		$new: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve();
			return deferred.promise();
		},
		$read: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve(_customers[op.$key]);
			return deferred.promise();
		},
		$query: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve({
				$resources: _customers
			});
			return deferred.promise();
		},
		$save: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve();
			return deferred.promise();
		},
		$delete: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve();
			return deferred.promise();
		}
	}
};
