// syracuse-tablet/html/js/demos/_entities/entCustomerRevenues.js
"use strict";

var _months = [
	"January",
	"February",
	"March",
	"April",
	"May",
	"June",
	"July",
	"August",
	"September",
	"October",
	"November",
	"December"
];

var _revs = [];
var _uuid = 0;
for (var c = 10000; c < 10100; c++) {
	for (var m = 0; m < 12; m++) {
		_revs.push({
			$uuid: "" + _uuid,
			bpcnum: "" + c,
			month: _months[m],
			revenue: 20000 + (m * (12 + (c % 12)) % 40) * 20000
		});

		_uuid++;
	}
}

exports.entity = {
	$entityName: "$localCustomerRevenue",
	$value: "{name}",
	$properties: {
		bpcnum: {
			$type: "application/x-string",
			$title: "ID"
		},
		month: {
			$type: "application/x-string",
			$title: "Month"
		},
		revenue: {
			$type: "application/x-decimal",
			$title: "Revenue"
		}
	},
	$prototype: function(proto) {
		// Add some chart information in $query prototype
		var res = proto.$properties && proto.$properties.$resources;
		if (res) {
			res.$axes = [{
				$title: "Month",
				$hierarchies: [
					["month"]
				]
			}, {
				$hierarchies: [
					["$measures"]
				]
			}];
			res.$cube = {
				$title: "Revenue",
				$hierarchies: {
					month: {
						$properties: ["month"]
					}
				},
				$measures: {
					revenue: {}
				}
			};
		}
	},
	$articles: {},
	$services: {
		$new: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve();
			return deferred.promise();
		},
		$read: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve(_revs[op.$key]);
			return deferred.promise();
		},
		$query: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve({
				$resources: _revs
			});
			return deferred.promise();
		},
		$save: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve();
			return deferred.promise();
		},
		$delete: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve();
			return deferred.promise();
		}
	}
};
