// syracuse-tablet/html/js/demos/_entities/entOverviewSales.js
"use strict";

var _data = {
	$uuid: "1",
	revMonthlyMin: 100000,
	revMonthly: 180000,
	revMonthlyMax: 200000,
	revMonthlyPerc: 92
};

exports.entity = {
	$entityName: "$localOverviewSales",
	$value: "{revMonthly}",
	$properties: {
		revMonthlyMin: {
			$type: "application/x-decimal",
			$title: "Min"
		},
		revMonthlyMax: {
			$type: "application/x-decimal",
			$title: "Target"
		},
		revMonthly: {
			$type: "application/x-decimal",
			$title: "Revenue"
		},
		revMonthlyPerc: {
			$type: "application/x-decimal",
			$title: "Monthly Target"
		},
	},
	$articles: {},
	$services: {
		$new: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve();
			return deferred.promise();
		},
		$read: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve(_data);
			return deferred.promise();
		},
		$query: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve({
				$resources: [_data]
			});
			return deferred.promise();
		},
		$save: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve();
			return deferred.promise();
		},
		$delete: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve();
			return deferred.promise();
		}
	}
};
