// syracuse-tablet/html/js/demos/_entities/entOverviewSalesLeads.js
"use strict";

var _data = [{
	$uuid: "1",
	date: "2014-12-05",
	numLeads: 50,
	numDeals: 40,
}, {
	$uuid: "1",
	date: "2014-12-12",
	numLeads: 53,
	numDeals: 45,
}, {
	$uuid: "1",
	date: "2014-12-15",
	numLeads: 54,
	numDeals: 47,
}, {
	$uuid: "1",
	date: "2014-12-22",
	numLeads: 53,
	numDeals: 49,
}, {
	$uuid: "1",
	date: "2014-12-29",
	numLeads: 52,
	numDeals: 52,
}];

exports.entity = {
	$entityName: "$localOverviewSalesLeads",
	$value: "{date}",
	$properties: {
		date: {
			$type: "application/x-date",
			$title: "Date"
		},
		numLeads: {
			$type: "application/x-decimal",
			$title: "Leads"
		},
		numDeals: {
			$type: "application/x-decimal",
			$title: "Deals"
		}
	},
	$articles: {},
	$prototype: function(proto) {

		// Add some chart information in $query prototype
		var res = proto.$properties && proto.$properties.$resources;
		if (res) {
			res.$axes = [{
				$title: "Date",
				$hierarchies: [
					["date"]
				]
			}, {
				$title: "Number",
				$hierarchies: [
					["$measures"]
				]
			}];
			res.$cube = {
				$style: "column",
				$title: "Leads vs Deals",
				$hierarchies: {
					date: {
						$properties: ["date"]
					}
				},
				$measures: {
					numLeads: {},
					numDeals: {}
				}
			};
		}
	},
	$services: {
		$new: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve();
			return deferred.promise();
		},
		$read: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve(_data[0]);
			return deferred.promise();
		},
		$query: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve({
				$resources: _data
			});
			return deferred.promise();
		},
		$save: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve();
			return deferred.promise();
		},
		$delete: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve();
			return deferred.promise();
		}
	}
};
