// syracuse-tablet/html/js/demos/_entities/entStatsVEN42.js
"use strict";

var _proto = {

	"$baseUrl": "/mobile1/x3/erp/SEEDTEST",
	"$url": "{$baseUrl}/STATS('$localVEN42')?representation=STATS~$localVEN42.$cube&role={$role}",
	"$title": "Sales by Region",
	"$type": "application/json; vnd.sage=syracuse.presentation=x3.erp.superv/STATS('$localVEN42')",
	"$methods": "GET",
	"codeStat": "$localVEN42",
	"$properties": {
		"datstart1": {
			"$type": "application/x-date",
			"$title": "Period start 1",
			"$isAdvanced": true,
			"$links": {
				"$details": {
					"$url": "{$baseUrl}/STATS('$localVEN42')?representation=STATS~$localVEN42.$cube&{linkstartP}",
					"$type": "application/json;vnd.sage=syracuse"
				}
			},
			"$isExcluded": true
		},
		"datstart2": {
			"$type": "application/x-date",
			"$title": "Period start 2",
			"$isAdvanced": true,
			"$links": {
				"$details": {
					"$url": "{$baseUrl}/STATS('VEN42')?representation=STATS~VEN42.$cube&{linkstartP}",
					"$type": "application/json;vnd.sage=syracuse"
				}
			},
			"$isExcluded": true
		},
		"linkstartP": {
			"$type": "application/x-string",
			"$isExcluded": true
		},
		"datend1": {
			"$type": "application/x-date",
			"$title": "Period end 1",
			"$isAdvanced": true,
			"$links": {
				"$details": {
					"$url": "{$baseUrl}/STATS('VEN42')?representation=STATS~VEN42.$cube&{linkendP}",
					"$type": "application/json;vnd.sage=syracuse"
				}
			},
			"$isExcluded": true
		},
		"datend2": {
			"$type": "application/x-date",
			"$title": "Period end 2",
			"$isAdvanced": true,
			"$links": {
				"$details": {
					"$url": "{$baseUrl}/STATS('VEN42')?representation=STATS~VEN42.$cube&{linkendP}",
					"$type": "application/json;vnd.sage=syracuse"
				}
			},
			"$isExcluded": true
		},
		"linkendP": {
			"$type": "application/x-string",
			"$isExcluded": true
		},
		"def_company": {
			"$type": "application/x-string",
			"$title": "Company",
			"$isAdvanced": true,
			"$links": {
				"$details": {
					"$type": "application/json;vnd.sage=syracuse",
					"$url": "{$baseUrl}/COMPANY('{def_company}')?representation=COMPANY.$details&role={$role}"
				}
			},
			"$isExcluded": true
		},
		"$resources": {
			"$type": "application/x-array",
			"$provType": "PORTALSTA",
			"$criteria": {
				"COMPANY": {
					"$type": "application/x-string",
					"$maxlength": 5,
					"$title": "Company",
					"$links": {
						"$details": {
							"$type": "application/json;vnd.sage=syracuse",
							"$url": "{$baseUrl}/COMPANY('{COMPANY}')?representation=COMPANY.$details&role={$role}",
							"$target": "main"
						}
					}
				},
				"DATBEG1": {
					"$type": "application/x-date",
					"$title": ""
				},
				"DATEND1": {
					"$type": "application/x-date",
					"$title": ""
				},
				"DATBEG2": {
					"$type": "application/x-date",
					"$title": ""
				},
				"DATEND2": {
					"$type": "application/x-date",
					"$title": ""
				}
			},
			"$item": {
				"$type": "application/json",
				"$url": "{$baseUrl}/stats('VEN42')?representation=$cube('VEN42')&role={$role}",
				"$title": "Sales by Region",
				"$properties": {
					"linkDown": {
						"$type": "application/x-string",
						"$isExcluded": true
					},
					"dateStat": {
						"$type": "application/x-date",
						"$capabilities": "sort,filter",
						"$isExcluded": true
					},
					"CRY": {
						"$title": "Pays",
						"$capabilities": "sort,filter",
						"$type": "application/x-string",
						"$links": {
							"$details": {
								"$type": "application/json;vnd.sage=syracuse",
								"$url": "{$baseUrl}/TABCOUNTRY('{CRY}')?representation=TABCOUNTRY.$details&role={$role}",
								"$target": "main"
							}
						}
					},
					"C_CRY": {
						"$capabilities": "sort,filter",
						"$type": "application/x-string"
					},
					"POSCOD": {
						"$title": "Code postal",
						"$capabilities": "sort,filter",
						"$type": "application/x-string",
						"$constraints": {
							"$maxLength": 3
						},
						"$isHidden": false,
						"$isParameter": true
					},
					"C_POSCOD": {
						"$capabilities": "sort,filter",
						"$type": "application/x-string"
					},
					"CTY": {
						"$title": "Ville",
						"$capabilities": "sort,filter",
						"$type": "application/x-string",
						"$constraints": {
							"$maxLength": 3
						},
						"$isHidden": false,
						"$isParameter": true
					},
					"C_CTY": {
						"$capabilities": "sort,filter",
						"$type": "application/x-string"
					},
					"BPCNUM": {
						"$title": "Client",
						"$capabilities": "sort,filter",
						"$type": "application/x-string",
						"$links": {
							"$details": {
								"$type": "application/json;vnd.sage=syracuse",
								"$url": "{$baseUrl}/BPARTNER('{BPCNUM}')?representation=BPARTNER.$details&role={$role}",
								"$target": "main"
							}
						}
					},
					"C_BPCNUM": {
						"$capabilities": "sort,filter",
						"$type": "application/x-string"
					},
					"amount1": {
						"$type": "application/x-decimal",
						"$precision": 15,
						"$scale": 2,
						"$format": "#,###.00",
						"$title": "Actual Year N"
					},
					"amount2": {
						"$type": "application/x-decimal",
						"$precision": 15,
						"$scale": 2,
						"$format": "#,###.00",
						"$title": "Actual Year N-1"
					},
					"amount3": {
						"$type": "application/x-decimal",
						"$precision": 15,
						"$scale": 2,
						"$format": "#,###.00",
						"$title": "Actual variance N and N-1"
					},
					"amount4": {
						"$type": "application/x-decimal",
						"$precision": 15,
						"$scale": 2,
						"$format": "#,###.00",
						"$title": "Margin Year N"
					},
					"amount5": {
						"$type": "application/x-decimal",
						"$precision": 15,
						"$scale": 2,
						"$format": "#,###.00",
						"$title": "Résult"
					},
					"amount6": {
						"$type": "application/x-decimal",
						"$precision": 15,
						"$scale": 2,
						"$format": "#,###.00",
						"$title": "Actual Tax Incl. Year N"
					},
					"amount7": {
						"$type": "application/x-decimal",
						"$precision": 15,
						"$scale": 2,
						"$format": "#,###.00",
						"$title": "Quantity"
					},
					"linkUp": {
						"$type": "application/x-string",
						"$isExcluded": true
					}
				},
				"$links": {}
			},
			"$cube": {
				"$mode": "table",
				"$title": "Inv. / Zone",
				"$displaysOneMeasure": true,
				"$hierarchies": {
					"CRY": {
						"$title": "Pays",
						"$properties": ["CRY"]
					},
					"POSCOD": {
						"$title": "Code postal",
						"$properties": ["POSCOD"]
					},
					"CTY": {
						"$title": "Ville",
						"$properties": ["CTY"]
					},
					"BPCNUM": {
						"$title": "Client",
						"$properties": ["BPCNUM"]
					},
					"date": {
						"properties": ["dateStat"],
						"periods": "YSQMD"
					}
				},
				"$measures": {
					"amount1": {
						"$isDefault": true
					},
					"amount2": {},
					"amount3": {},
					"amount4": {},
					"amount5": {},
					"amount6": {},
					"amount7": {}
				}
			},
			"$axes": [{
				"$title": "Pays",
				"$hierarchies": [
					["CRY", 1, 0],
					["POSCOD", 0, 0],
					["CTY", 0, 0],
					["BPCNUM", 0, 0]
				]
			}, {
				"$title": "Measures",
				"$hierarchies": [
					["$measures"]
				]
			}],
			"$slicer": {
				"$title": "CRY",
				"$hierarchies": [
					["CRY", 1, 0]
				]
			},
			"$isExcluded": false
		}
	}

};
var _countries = [{
	name: 'Afghanistan',
	code: 'AF'
}, {
	name: 'Åland Islands',
	code: 'AX'
}, {
	name: 'Albania',
	code: 'AL'
}, {
	name: 'Algeria',
	code: 'DZ'
}, {
	name: 'American Samoa',
	code: 'AS'
}, {
	name: 'AndorrA',
	code: 'AD'
}, {
	name: 'Angola',
	code: 'AO'
}, {
	name: 'Anguilla',
	code: 'AI'
}, {
	name: 'Antarctica',
	code: 'AQ'
}, {
	name: 'Antigua and Barbuda',
	code: 'AG'
}, {
	name: 'Argentina',
	code: 'AR'
}, {
	name: 'Armenia',
	code: 'AM'
}, {
	name: 'Aruba',
	code: 'AW'
}, {
	name: 'Australia',
	code: 'AU'
}, {
	name: 'Austria',
	code: 'AT'
}, {
	name: 'Azerbaijan',
	code: 'AZ'
}, {
	name: 'Bahamas',
	code: 'BS'
}, {
	name: 'Bahrain',
	code: 'BH'
}, {
	name: 'Bangladesh',
	code: 'BD'
}, {
	name: 'Barbados',
	code: 'BB'
}, {
	name: 'Belarus',
	code: 'BY'
}, {
	name: 'Belgium',
	code: 'BE'
}, {
	name: 'Belize',
	code: 'BZ'
}, {
	name: 'Benin',
	code: 'BJ'
}, {
	name: 'Bermuda',
	code: 'BM'
}, {
	name: 'Bhutan',
	code: 'BT'
}, {
	name: 'Bolivia',
	code: 'BO'
}, {
	name: 'Bosnia and Herzegovina',
	code: 'BA'
}, {
	name: 'Botswana',
	code: 'BW'
}, {
	name: 'Bouvet Island',
	code: 'BV'
}, {
	name: 'Brazil',
	code: 'BR'
}, {
	name: 'British Indian Ocean Territory',
	code: 'IO'
}, {
	name: 'Brunei Darussalam',
	code: 'BN'
}, {
	name: 'Bulgaria',
	code: 'BG'
}, {
	name: 'Burkina Faso',
	code: 'BF'
}, {
	name: 'Burundi',
	code: 'BI'
}, {
	name: 'Cambodia',
	code: 'KH'
}, {
	name: 'Cameroon',
	code: 'CM'
}, {
	name: 'Canada',
	code: 'CA'
}, {
	name: 'Cape Verde',
	code: 'CV'
}, {
	name: 'Cayman Islands',
	code: 'KY'
}, {
	name: 'Central African Republic',
	code: 'CF'
}, {
	name: 'Chad',
	code: 'TD'
}, {
	name: 'Chile',
	code: 'CL'
}, {
	name: 'China',
	code: 'CN'
}, {
	name: 'Christmas Island',
	code: 'CX'
}, {
	name: 'Cocos (Keeling) Islands',
	code: 'CC'
}, {
	name: 'Colombia',
	code: 'CO'
}, {
	name: 'Comoros',
	code: 'KM'
}, {
	name: 'Congo',
	code: 'CG'
}, {
	name: 'Congo, The Democratic Republic of the',
	code: 'CD'
}, {
	name: 'Cook Islands',
	code: 'CK'
}, {
	name: 'Costa Rica',
	code: 'CR'
}, {
	name: 'Cote D\'Ivoire',
	code: 'CI'
}, {
	name: 'Croatia',
	code: 'HR'
}, {
	name: 'Cuba',
	code: 'CU'
}, {
	name: 'Cyprus',
	code: 'CY'
}, {
	name: 'Czech Republic',
	code: 'CZ'
}, {
	name: 'Denmark',
	code: 'DK'
}, {
	name: 'Djibouti',
	code: 'DJ'
}, {
	name: 'Dominica',
	code: 'DM'
}, {
	name: 'Dominican Republic',
	code: 'DO'
}, {
	name: 'Ecuador',
	code: 'EC'
}, {
	name: 'Egypt',
	code: 'EG'
}, {
	name: 'El Salvador',
	code: 'SV'
}, {
	name: 'Equatorial Guinea',
	code: 'GQ'
}, {
	name: 'Eritrea',
	code: 'ER'
}, {
	name: 'Estonia',
	code: 'EE'
}, {
	name: 'Ethiopia',
	code: 'ET'
}, {
	name: 'Falkland Islands (Malvinas)',
	code: 'FK'
}, {
	name: 'Faroe Islands',
	code: 'FO'
}, {
	name: 'Fiji',
	code: 'FJ'
}, {
	name: 'Finland',
	code: 'FI'
}, {
	name: 'France',
	code: 'FR'
}, {
	name: 'French Guiana',
	code: 'GF'
}, {
	name: 'French Polynesia',
	code: 'PF'
}, {
	name: 'French Southern Territories',
	code: 'TF'
}, {
	name: 'Gabon',
	code: 'GA'
}, {
	name: 'Gambia',
	code: 'GM'
}, {
	name: 'Georgia',
	code: 'GE'
}, {
	name: 'Germany',
	code: 'DE'
}, {
	name: 'Ghana',
	code: 'GH'
}, {
	name: 'Gibraltar',
	code: 'GI'
}, {
	name: 'Greece',
	code: 'GR'
}, {
	name: 'Greenland',
	code: 'GL'
}, {
	name: 'Grenada',
	code: 'GD'
}, {
	name: 'Guadeloupe',
	code: 'GP'
}, {
	name: 'Guam',
	code: 'GU'
}, {
	name: 'Guatemala',
	code: 'GT'
}, {
	name: 'Guernsey',
	code: 'GG'
}, {
	name: 'Guinea',
	code: 'GN'
}, {
	name: 'Guinea-Bissau',
	code: 'GW'
}, {
	name: 'Guyana',
	code: 'GY'
}, {
	name: 'Haiti',
	code: 'HT'
}, {
	name: 'Heard Island and Mcdonald Islands',
	code: 'HM'
}, {
	name: 'Holy See (Vatican City State)',
	code: 'VA'
}, {
	name: 'Honduras',
	code: 'HN'
}, {
	name: 'Hong Kong',
	code: 'HK'
}, {
	name: 'Hungary',
	code: 'HU'
}, {
	name: 'Iceland',
	code: 'IS'
}, {
	name: 'India',
	code: 'IN'
}, {
	name: 'Indonesia',
	code: 'ID'
}, {
	name: 'Iran, Islamic Republic Of',
	code: 'IR'
}, {
	name: 'Iraq',
	code: 'IQ'
}, {
	name: 'Ireland',
	code: 'IE'
}, {
	name: 'Isle of Man',
	code: 'IM'
}, {
	name: 'Israel',
	code: 'IL'
}, {
	name: 'Italy',
	code: 'IT'
}, {
	name: 'Jamaica',
	code: 'JM'
}, {
	name: 'Japan',
	code: 'JP'
}, {
	name: 'Jersey',
	code: 'JE'
}, {
	name: 'Jordan',
	code: 'JO'
}, {
	name: 'Kazakhstan',
	code: 'KZ'
}, {
	name: 'Kenya',
	code: 'KE'
}, {
	name: 'Kiribati',
	code: 'KI'
}, {
	name: 'Korea, Democratic People\'S Republic of',
	code: 'KP'
}, {
	name: 'Korea, Republic of',
	code: 'KR'
}, {
	name: 'Kuwait',
	code: 'KW'
}, {
	name: 'Kyrgyzstan',
	code: 'KG'
}, {
	name: 'Lao People\'S Democratic Republic',
	code: 'LA'
}, {
	name: 'Latvia',
	code: 'LV'
}, {
	name: 'Lebanon',
	code: 'LB'
}, {
	name: 'Lesotho',
	code: 'LS'
}, {
	name: 'Liberia',
	code: 'LR'
}, {
	name: 'Libyan Arab Jamahiriya',
	code: 'LY'
}, {
	name: 'Liechtenstein',
	code: 'LI'
}, {
	name: 'Lithuania',
	code: 'LT'
}, {
	name: 'Luxembourg',
	code: 'LU'
}, {
	name: 'Macao',
	code: 'MO'
}, {
	name: 'Macedonia, The Former Yugoslav Republic of',
	code: 'MK'
}, {
	name: 'Madagascar',
	code: 'MG'
}, {
	name: 'Malawi',
	code: 'MW'
}, {
	name: 'Malaysia',
	code: 'MY'
}, {
	name: 'Maldives',
	code: 'MV'
}, {
	name: 'Mali',
	code: 'ML'
}, {
	name: 'Malta',
	code: 'MT'
}, {
	name: 'Marshall Islands',
	code: 'MH'
}, {
	name: 'Martinique',
	code: 'MQ'
}, {
	name: 'Mauritania',
	code: 'MR'
}, {
	name: 'Mauritius',
	code: 'MU'
}, {
	name: 'Mayotte',
	code: 'YT'
}, {
	name: 'Mexico',
	code: 'MX'
}, {
	name: 'Micronesia, Federated States of',
	code: 'FM'
}, {
	name: 'Moldova, Republic of',
	code: 'MD'
}, {
	name: 'Monaco',
	code: 'MC'
}, {
	name: 'Mongolia',
	code: 'MN'
}, {
	name: 'Montserrat',
	code: 'MS'
}, {
	name: 'Morocco',
	code: 'MA'
}, {
	name: 'Mozambique',
	code: 'MZ'
}, {
	name: 'Myanmar',
	code: 'MM'
}, {
	name: 'Namibia',
	code: 'NA'
}, {
	name: 'Nauru',
	code: 'NR'
}, {
	name: 'Nepal',
	code: 'NP'
}, {
	name: 'Netherlands',
	code: 'NL'
}, {
	name: 'Netherlands Antilles',
	code: 'AN'
}, {
	name: 'New Caledonia',
	code: 'NC'
}, {
	name: 'New Zealand',
	code: 'NZ'
}, {
	name: 'Nicaragua',
	code: 'NI'
}, {
	name: 'Niger',
	code: 'NE'
}, {
	name: 'Nigeria',
	code: 'NG'
}, {
	name: 'Niue',
	code: 'NU'
}, {
	name: 'Norfolk Island',
	code: 'NF'
}, {
	name: 'Northern Mariana Islands',
	code: 'MP'
}, {
	name: 'Norway',
	code: 'NO'
}, {
	name: 'Oman',
	code: 'OM'
}, {
	name: 'Pakistan',
	code: 'PK'
}, {
	name: 'Palau',
	code: 'PW'
}, {
	name: 'Palestinian Territory, Occupied',
	code: 'PS'
}, {
	name: 'Panama',
	code: 'PA'
}, {
	name: 'Papua New Guinea',
	code: 'PG'
}, {
	name: 'Paraguay',
	code: 'PY'
}, {
	name: 'Peru',
	code: 'PE'
}, {
	name: 'Philippines',
	code: 'PH'
}, {
	name: 'Pitcairn',
	code: 'PN'
}, {
	name: 'Poland',
	code: 'PL'
}, {
	name: 'Portugal',
	code: 'PT'
}, {
	name: 'Puerto Rico',
	code: 'PR'
}, {
	name: 'Qatar',
	code: 'QA'
}, {
	name: 'Reunion',
	code: 'RE'
}, {
	name: 'Romania',
	code: 'RO'
}, {
	name: 'Russian Federation',
	code: 'RU'
}, {
	name: 'RWANDA',
	code: 'RW'
}, {
	name: 'Saint Helena',
	code: 'SH'
}, {
	name: 'Saint Kitts and Nevis',
	code: 'KN'
}, {
	name: 'Saint Lucia',
	code: 'LC'
}, {
	name: 'Saint Pierre and Miquelon',
	code: 'PM'
}, {
	name: 'Saint Vincent and the Grenadines',
	code: 'VC'
}, {
	name: 'Samoa',
	code: 'WS'
}, {
	name: 'San Marino',
	code: 'SM'
}, {
	name: 'Sao Tome and Principe',
	code: 'ST'
}, {
	name: 'Saudi Arabia',
	code: 'SA'
}, {
	name: 'Senegal',
	code: 'SN'
}, {
	name: 'Serbia and Montenegro',
	code: 'CS'
}, {
	name: 'Seychelles',
	code: 'SC'
}, {
	name: 'Sierra Leone',
	code: 'SL'
}, {
	name: 'Singapore',
	code: 'SG'
}, {
	name: 'Slovakia',
	code: 'SK'
}, {
	name: 'Slovenia',
	code: 'SI'
}, {
	name: 'Solomon Islands',
	code: 'SB'
}, {
	name: 'Somalia',
	code: 'SO'
}, {
	name: 'South Africa',
	code: 'ZA'
}, {
	name: 'South Georgia and the South Sandwich Islands',
	code: 'GS'
}, {
	name: 'Spain',
	code: 'ES'
}, {
	name: 'Sri Lanka',
	code: 'LK'
}, {
	name: 'Sudan',
	code: 'SD'
}, {
	name: 'Suriname',
	code: 'SR'
}, {
	name: 'Svalbard and Jan Mayen',
	code: 'SJ'
}, {
	name: 'Swaziland',
	code: 'SZ'
}, {
	name: 'Sweden',
	code: 'SE'
}, {
	name: 'Switzerland',
	code: 'CH'
}, {
	name: 'Syrian Arab Republic',
	code: 'SY'
}, {
	name: 'Taiwan, Province of China',
	code: 'TW'
}, {
	name: 'Tajikistan',
	code: 'TJ'
}, {
	name: 'Tanzania, United Republic of',
	code: 'TZ'
}, {
	name: 'Thailand',
	code: 'TH'
}, {
	name: 'Timor-Leste',
	code: 'TL'
}, {
	name: 'Togo',
	code: 'TG'
}, {
	name: 'Tokelau',
	code: 'TK'
}, {
	name: 'Tonga',
	code: 'TO'
}, {
	name: 'Trinidad and Tobago',
	code: 'TT'
}, {
	name: 'Tunisia',
	code: 'TN'
}, {
	name: 'Turkey',
	code: 'TR'
}, {
	name: 'Turkmenistan',
	code: 'TM'
}, {
	name: 'Turks and Caicos Islands',
	code: 'TC'
}, {
	name: 'Tuvalu',
	code: 'TV'
}, {
	name: 'Uganda',
	code: 'UG'
}, {
	name: 'Ukraine',
	code: 'UA'
}, {
	name: 'United Arab Emirates',
	code: 'AE'
}, {
	name: 'United Kingdom',
	code: 'GB'
}, {
	name: 'United States',
	code: 'US'
}, {
	name: 'United States Minor Outlying Islands',
	code: 'UM'
}, {
	name: 'Uruguay',
	code: 'UY'
}, {
	name: 'Uzbekistan',
	code: 'UZ'
}, {
	name: 'Vanuatu',
	code: 'VU'
}, {
	name: 'Venezuela',
	code: 'VE'
}, {
	name: 'Viet Nam',
	code: 'VN'
}, {
	name: 'Virgin Islands, British',
	code: 'VG'
}, {
	name: 'Virgin Islands, U.S.',
	code: 'VI'
}, {
	name: 'Wallis and Futuna',
	code: 'WF'
}, {
	name: 'Western Sahara',
	code: 'EH'
}, {
	name: 'Yemen',
	code: 'YE'
}, {
	name: 'Zambia',
	code: 'ZM'
}, {
	name: 'Zimbabwe',
	code: 'ZW'
}];

var _resources = [];
_countries
	.forEach(function(ctry, idx) {
		_resources
			.push({
				"level": 1,
				"CRY": ctry.code,
				"$uuid": "uuid" + ctry.code,
				"$title": ctry.name,
				"linkUp": "DATBEG1=2015-01-01&DATEND1=2015-12-31&DATBEG2=2014-01-01&DATEND2=2014-12-31&COMPANY=&CRY=" + ctry.code,
				"linkDown": "DATBEG1=2015-01-01&DATEND1=2015-12-31&DATBEG2=2014-01-01&DATEND2=2014-12-31&COMPANY=",
				"amount1": 1000 + idx * 10,
				"amount2": 2000 + idx * 10,
				"amount3": 3000 + idx * 10,
				"amount4": 4000 + idx * 10,
				"amount5": 5000 + idx * 10,
				"amount6": 6000 + idx * 10,
				"amount7": 7000 + idx * 10,
			});
	});
// test pagination - lastIndex
// _resources.splice (_resources.length - 4, 3)
// _resources.splice (_resources.length - 3, 2)
// _resources.splice (_resources.length - 2, 1)
var _data = {
	"$properties": {
		"$resources": {
			"$item": {
				"$properties": {
					"CRY": {
						"$isHidden": false
					},
					"POSCOD": {
						"$isHidden": true
					},
					"CTY": {
						"$isHidden": true
					},
					"BPCNUM": {
						"$isHidden": true
					}
				},
				"$links": {}
			}
		}
	},
	"linkstartP": "DATBEG1=2014-01-01&DATEND1=2014-12-31&DATBEG2=2013-01-01&DATEND2=2013-12-31&COMPANY=",
	"linkendP": "DATBEG1=2016-01-01&DATEND1=2016-12-31&DATBEG2=2015-01-01&DATEND2=2015-12-31&COMPANY=",
	"datstart1": "2015-01-01",
	"datend1": "2015-12-31",
	"datstart2": "2014-01-01",
	"datend2": "2014-12-31",
	"def_company": "",
	"$resources": _resources
};

var _chartDetail = {
	"$layoutType": "stack",
	"$display": "bothSideToSide",
	"$localization": {},
	"$items": [{
		"$layoutType": "row",
		"$items": [{
			"$layoutType": "cell",
			"widthAll": "6",
			"$widthSm": "6",
			"$widthMd": "6",
			"$widthLg": "6",
			"$bgColor": "transparent",
			"$items": [{
				"$bind": "$resources",
				"$isTitleHidden": true,
				"$display": "chart",
				"$isHidden": false,
				"$style": "line",
				"$arrayOptions": {
					"showPagination": false,
					"showSortFilter": false,
					"showSearch": false,
					"rowIndex": false
				},
				"$arrayBgColor": "darkgrey"
			}]
		}, {
			"$layoutType": "cell",
			"$widthXs": "12",
			"$widthSm": "6",
			"$widthMd": "6",
			"$widthLg": "6",
			"$bgColor": "transparent",
			"$items": [{
				"$bind": "$resources",
				"$isTitleHidden": true,
				"$display": "table",
				"$isHidden": false,
				"$arrayOptions": {
					"showPagination": true,
					"showSortFilter": false,
					"showSearch": false,
					"rowIndex": true,
					"showRowDetailLink": false
				},
				"$arrayBgColor": "darkgrey",
				"$table": [{
					"$bind": "CRY",
					"$title": "Pays",
					"$type": "application/x-string",
					"$titleOrig": "Pays",
					"$widthType": "auto",
					"$widthVal": 0,
					"$imgHeight": "small"
				}, {
					"$bind": "amount1",
					"$title": "Actual Year N",
					"$type": "application/x-decimal",
					"$titleOrig": "Actual Year N",
					"$widthType": "auto",
					"$widthVal": 0,
					"$imgHeight": "small"
				}, {
					"$bind": "amount2",
					"$title": "Actual Year N-1",
					"$type": "application/x-decimal",
					"$titleOrig": "Actual Year N-1",
					"$widthType": "auto",
					"$widthVal": 0,
					"$imgHeight": "small"
				}, {
					"$bind": "amount3",
					"$title": "Actual variance N and N-1",
					"$type": "application/x-decimal",
					"$titleOrig": "Actual variance N and N-1",
					"$widthType": "auto",
					"$widthVal": 0,
					"$imgHeight": "small"
				}, {
					"$bind": "amount4",
					"$title": "Margin Year N",
					"$type": "application/x-decimal",
					"$titleOrig": "Margin Year N",
					"$widthType": "auto",
					"$widthVal": 0,
					"$imgHeight": "small"
				}]
			}]
		}]
	}]
};

exports.entity = {
	$entityName: "$localVEN42",
	$value: "{date}",
	$proto: _proto,
	$createProto: function(prototype) {
		return $.extend(true, {}, exports.entity.$proto);
	},
	$properties: {},
	$articles: {
		$cube: {
			"$layoutType": "stack",
			"$items": [{
				"$bind": "$resources",
				"$display": "table",
				"$arrayOptions": {
					"showPagination": true,
					"rowIndex": true,
				},
				"$table": [{
					"$bind": "CRY"
				}, {
					"$bind": "amount1"
				}, {
					"$bind": "amount2"
				}, {
					"$bind": "amount3"
				}],
				/*
				 * { "$bind": "amount4" }, { "$bind": "amount5" }, {
				 * "$bind": "amount6" }, { "$bind": "amount7" }],
				 */
				$chartDetail: _chartDetail
			}]
		},
		$cubechart: {
			"$layoutType": "stack",
			"$items": [{
				"$bind": "$resources",
				"$display": "chart",
				"$arrayOptions": {
					"showPagination": true,
					"rowIndex": true,
				},
				"$chartOptions": {
					"isLegendHidden": true
				},
				"$table": [{
					"$bind": "CRY"
				}, {
					"$bind": "amount1"
				}, {
					"$bind": "amount2"
				}, {
					"$bind": "amount3"
				}],
				$chartDetail: _chartDetail
			}]
		}
	},
	$services: {
		$query: function(op, data) {
			return $.smResolve($.extend(true, {}, _data));
		}
	}
};
