// syracuse-ui/lib/site/favorites/favoriteMenuItem.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var ReferenceField = require('syracuse-ui/lib/field/referenceField').ReferenceField;

var _$executes = ["$execute_$query", "$execute_$details", "$execute_$edit"];

function FavoriteMenuItem() {}

exports.FavoriteMenuItem = helpers.defineClass(FavoriteMenuItem, ReferenceField, {
	loadBox: function() {
		ReferenceField.prototype.loadBox.call(this);
	},
	render: function() {
		var css = ((this.$item.$css) ? this.$item.$css + " " : "");
		var $skinInput = css + this.$skinInput;
		this._input = document.createElement("input");
		// add attribute to disable browser spellcheck and autocorrect
		document.site.setSpecificAttributes(this._input);
		this._input.setAttribute("type", "text");
		this.$$input = $(this._input);
		this._input.className = $skinInput;

		this.fieldValue.appendChild(this._input);

		this.tunnelPicker = this._appendSpecialPicker("tunnel");
		this.tunnelPicker.setAttribute("title", this.localize.fieldTunnel);

		this.setDescription(this.$item.description || this.$field.$item.description);

		this.$item.$url = this.articleParent.dataset.$singleField.menuItem.$url;
		if (this.articleParent.dataset.$singleField.menuItem.params) {
			this.$item.params = this.articleParent.dataset.$singleField.menuItem.params;
		}

		if (this.currentValue) {
			this._renderCurrentValue();
		}
	},
	onClickPicker: function(target, event) {
		if (target.syraPickerType == "tunnel") {
			this.page.toggle(false);
			document.site.favoritesProvider.navigate(this);
			return false;
		}
		return true;
	}

});
