// syracuse-ui/lib/site/favorites/favorites.js
"use strict";
/*
 Copyright (c)2012 Sage.
 http://wwww.sage.com
 */
function _getPrototype(localize) {
	return {
		$type: "application/json",
		$title: "Favoris",
		$properties: {
			items: {
				$type: "application/x-array",
				$capabilities: "reorder",
				$item: {
					$type: "application/x-variant",
					$variants: {
						menuItem: {
							$type: "application/x-reference",
							$capabilities: "delete,reorder",
							$item: {
								$value: "{title}",
								$key: "{$uuid}",
								$properties: {
									title: {
										$type: "application/x-string"
									}
								}
							}
						},
						menuBlock: {
							$type: "application/x-object",
							$capabilities: "delete",
							$item: {
								$value: "{title}",
								$key: "{$uuid}",
								$properties: {
									title: {
										$type: "application/x-string"
									},
									items: {
										$type: "application/x-array",
										$capabilities: "reorder",
										$item: {
											$type: "application/x-variant",
											$variants: {
												menuItem: {
													$type: "application/x-reference",
													$capabilities: "delete",
													$item: {
														$value: "{title}",
														$key: "{$uuid}",
														$properties: {
															title: {
																$type: "application/x-string"
															}
														}
													}
												},
												menuBlock: {
													$type: "application/x-object",
													$capabilities: "delete",
													$item: {
														$type: "application/x-pointer"
													}
												}
											}

										}

									}
								}
							}
						}
					}
				}
			}
		}
	};

};

function _getArticle(localize) {
	return {
		$layout: {
			$items: [{
				$category: "section",
				$layout: {
					$items: [{
						$category: "block",
						$layout: {
							$items: [{
								$isEditMode: true,
								$bind: "items",
								$isQuickDesignerEnabled: false,
								$skin: "s-favorites",
								$layout: {
									$items: [{
										$bind: "menuItem"
									}]
								}
							}]
						}
					}]
				}
			}]
		}
	};
};
exports.getArticle = function() {
	return _getArticle(document.site.localize);
};
exports.getPrototype = function() {
	return _getPrototype(document.site.localize);
};
exports.getFavoritesPage = function() {
	var localize = document.site.localize;
	var $prototype = _getPrototype(localize);
	$prototype.$article = _getArticle(localize);
	return $prototype;
};
