// syracuse-ui/lib/site/favorites/favoritesProvider.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function FavoritesProvider() {}

exports.FavoritesProvider = helpers.defineClass(FavoritesProvider, null, {
	save: function($url, $description) {
		if ($url) {
			if (!this.map) {
				this.map = {};
			}
			var start = $url.indexOf("?");
			if (start) {
				$url = $url.slice(start);
			}
			this.map[$url] = {
				$url: $url,
				$description: $description
			};
		}
	},
	onClickPicker: function(target, event) {
		if (target.syraMenuItemPath) {
			var field = document.controller.findField(target);
			document.site.toggleClass(target, "s-favorite-on", field.isFavorite = !field.isFavorite);
			if (field.isFavorite) {
				this.save(field.codeMenu.href);
			} else {
				delete this.map[field.codeMenu.href];
			}
		}
	},
	addFavoritePicker: function(item) {
		if (item.isNavigationMenuItemField) {
			if (item.codeMenu) {
				item.favoritePicker = document.createElement("a");
				item.favoritePicker.setAttribute("data-s-picker", item.favoritePicker.syraPickerType = "favorite");
				item.favoritePicker.syra_field_id = item.id;
				item.favoritePicker.syraMenuItemPath = item.page.getNavigationPath(item);
				item.favoritePicker.className = "s-nav-menu-favorite";
				item.fieldValue.parentNode.insertBefore(item.favoritePicker, item.fieldValue.nextSibling);
				if (item.codeMenu.href) {
					if (this.map && this.map[item.codeMenu.href]) {
						document.site.toggleClass(item.favoritePicker, "s-favorite-on", item.isFavorite = true);
					}
				} else {
					//document.site.toggleClass(item.layoutSlot, "s-favorite-off", true);
				}
			}
		}
	},
	dispose: function() {
		this.map = null;
	}
});
