// syracuse-ui/lib/field/array/gridBuilder.js
"use strict";
var helpers = require('syracuse-core').helpers;
var _gridCardBuilder = require('./gridMode/gridCardBuilder');
var _gridFilter = require('syracuse-ui/lib/field/filter/gridFilter');
var _colResizer = require("./gridMode/tools/colResizer");
var _colsReorderer = require('./gridMode/tools/colsReorderer');
var _popupCardBuilder = require('./gridMode/popupCardBuilder');
var _gridTable = require('./gridMode/gridTable');


function _appendRowIndexCell(record, row) {
	var btn = record.rowIndexBtn = document.createElement("td");
	var css = record.list.cssRowIndex;
	if (record.list.$item.$selectByRowIndex) {
		css += " s-list-row-index-selector";
		btn.syraOnClick = "onSelectorClick";
	}
	if (record.list.$capability.reorder) {
		record.reorderBtn = record.rowIndexBtn;
		record.reorderBtn.syraReorderRecordId = record.list.id;
		css += " s-grid-record-reorder";
		record.rowIndexBtn.syraDragSpot = record.id;
	}
	btn.className = css;
	row.appendChild(btn);
	record.setRowIndex();
	return btn;
}

function _initializeFusionNavigationList(list) {
	list.$fitContainer = list.$item.$isNavigationList;
	list.isScrollButtonEnabled = false;
	list.$item.$isPickerMenuHidden = true;
	list.$isQuickDesignerDisabled = true;
	list.$item.$isMenuRecordHidden = true;
	list.isClientFetch = false;
	list.isClientSortEnable = true;
	list.isClientPagingPerPageDisable = true;
	list.discardFieldsGarbage = true;
	list.focus = function(select, name, $serverIndex) {
		var bar = this.page && this.page.fusionBar;
		return bar && name ? bar.focus(select, name) : true;
	};
}

function _addFilterRecord(builder) {
	if (builder.list.$item.$isQuickFilter) {
		if (builder.filterRecord) {
			builder.filterRecord.drawRecord();
		} else {
			builder.filterRecord = _gridFilter.addRecord(builder.list);
		}
	}
}


function _addTitleRow(builder) {
	var list = builder.list;
	builder.allColumns = [];
	var $items = list.$item.$layout.$items;

	var fixedTable = builder.scrollTable;
	if (list.treeDecorator) {
		fixedTable = builder.scrollTable;
		fixedTable.addTreeCol();
	} else {
		var hasFixedCol = list.hasRowIndexCol = list.$item.$isRowIndexVisible || list.$capability.reorder || list.$item.$selectByRowIndex;
		for (var ii = 0, jj = $items.length; ii < jj; ii++) {
			var $item = $items[ii];
			var $field = list.$fields[$item.$bind];
			if ($field && !$field.$isExcluded) {
				if ($field.$capabilities && $field.$capabilities.indexOf("frozen") >= 0) {
					$field.$isFixed = true;
				}
				hasFixedCol = hasFixedCol || (($item.$isFixed !== undefined) ? $item.$isFixed : $field.$isFixed);
			}
		}
		if (hasFixedCol && !syra_quality.enableTestRobot && list.isFixedColEnabled) {
			fixedTable = builder.fixedTable = _gridTable.addFixed(builder);
			builder.scrollTable.slot.parentNode.insertBefore(fixedTable.slot, builder.scrollTable.slot);
		}
	}
	_addFilterRecord(builder);
	!list.treeDecorator && list.hasRowIndexCol && fixedTable.addRowIndexCol();
	_gridCardBuilder.addRowCardCol(builder);
	list.selector.useColumnSelector && fixedTable.addSelectorColumn();
	!list.$item.$isMenuRecordHidden && fixedTable.addMenusCol();

	var isScrollable;
	for (var ii = 0, jj = $items.length; ii < jj; ii++) {
		var $item = $items[ii];
		var $field = list.$fields[$item.$bind];
		if ($field && !$field.$isExcluded) {
			var $isFixed = !isScrollable && (($item.$isFixed !== undefined) ? $item.$isFixed : $field.$isFixed);
			if (!$isFixed) {
				isScrollable = true;
			}
			var col = ($isFixed ? fixedTable : builder.scrollTable).addFieldCol($item, $field);
			builder.filterRecord && builder.filterRecord.addFieldCell(col, $field);
		}
	}
	builder.filterRecord && builder.filterRecord.validateDisplay();
	_checkHeaderVisibility(builder);
}

function _checkHeaderVisibility(builder) {
	var isTitleRowVisible = false,
		isFilterVisible = false;
	for (var ii = 0, jj = builder.allColumns.length; ii < jj; ii++) {
		var col = builder.allColumns[ii];
		if (col.$field && !col.$field.$isExcluded) {
			if (!isTitleRowVisible && col.titleText) {
				isTitleRowVisible = true;
			}
			if (!col.$isHidden && col.$field.$hasFilterCapability) {
				isFilterVisible = true;
			}
		}
	}
	var filter = builder.filterRecord;
	if (filter && filter.isFilterVisible !== isFilterVisible) {
		filter.isFilterVisible = isFilterVisible;
		syra_dom.hide(filter.fixedRow, !isFilterVisible);
		syra_dom.hide(filter.dataRow, !isFilterVisible);
	}
	if (builder.isTitleRowVisible !== isTitleRowVisible) {
		builder.isTitleRowVisible = isTitleRowVisible;
		builder.fixedTable && syra_dom.hide(builder.fixedTable.headSlot, !isTitleRowVisible);
		syra_dom.hide(builder.scrollTable.headSlot, !isTitleRowVisible);
	}
}

function _addRow(builder, record) {
	var row = document.createElement("tr");
	row.syraItem = row.syrainout = record.id;
	var css = record.list.$skin + "-row";
	if (record.list.$item.$cssRow) {
		css += " " + record.list.$item.$cssRow;
	}
	row.className = css;
	if (record.list.selector.isRowMode || builder.isOutCardMode) {
		row.syraOnClick = "onSelectorClick";
		row.className += " s-list-selector-row";
	}
	return row;
}

function GridBuilder() {}

exports.GridBuilder = helpers.defineClass(GridBuilder, null, {
	initialize: function(list) {
		this.list = list;
		this.list.enableDefaultDesigner();
		list.$field.$isFusionNavigationList && _initializeFusionNavigationList(list);
		this.columnsMap = {};
		this.allColumns = [];
		list.onOverResize = function(additionalHeight) {
			var bodySlot = this.builder.scrollTable && this.builder.scrollTable.bodySlot;
			if (bodySlot) {
				var diffH = bodySlot.scrollHeight - (bodySlot.clientHeight + 1);
				return diffH > 0 ? Math.min(diffH, additionalHeight) : 0;
			}
		};
		list.emptyBody = function() {
			this.builder.fixedTable && this.builder.fixedTable.empty();
			this.builder.scrollTable.empty();
		};
		list.showEmptyItem = function(show) {
			if (show) {
				this.builder.scrollTable.bodySlot.appendChild(this.builder.emptyDataItem);
			} else {
				syra_dom.remove(this.builder.emptyDataItem);
			}
		};
		list.scrollToRecordField = function(field) {
			var col = this.builder.columnsMap[field.$item.$bind];
			if (col && !col.table.isFixedCols) {
				this.page.scrollToItem(field.domItem, col.table.bodySlot, true);
			}
		};
		_popupCardBuilder.load(this);
	},
	showBody: function(show) {
		this.fixedTable && this.fixedTable.showBody(show);
		this.scrollTable.showBody(show);
	},
	findEventCol: function(event) {
		var target = event.target;
		while (!target.syra_field_bind) {
			target = target.parentNode;
		}
		return this.columnsMap[target.syra_field_bind];
	},
	endInitiallize: function() {
		this.list.$defaultSkin = "s-" + this.list.$item.$format;
		this.list.$item.$isTopLabelAlignment = true;

		this.list.$complexBinds = [];
		var $colBinds = [];
		var defined = [];
		var $binds = Object.keys(this.list.$fields);
		for (var ii = 0, jj = $binds.length; ii < jj; ii++) {
			var $bind = $binds[ii];
			var $field = this.list.$fields[$bind];
			if (!$field.$isExcluded && !($field.$viewScope && $field.$viewScope == "popup")) {
				switch ($field.$type) {
					case "application/x-variant":
						//need support for later
						break;
					case "application/x-tag-cloud":
					case "application/x-array":
					case "application/x-object":
					case "text/rtf":
					case "text/html":
					case "text/plain":
						this.list.$complexBinds.push($bind);
						break;
					default:
						$colBinds.push($bind);
						defined.push($bind);
						break;
				}
			}
		}

		if (!this.list.$item.$layout) {
			this.list.$item.$layout = (this.list.$designing && this.list.$designing.grid && this.list.$designing.grid.$layout) || {};
		}
		//define default columns

		var $items = this.list.$item.$layout.$items;
		if (!$items) {
			$items = this.list.$item.$layout.$items = [];
			for (var ii = 0, jj = $colBinds.length; ii < jj; ii++) {
				$items.push({
					$bind: $colBinds[ii]
				});
			}
		} else {
			//check items for having only colbinds
			if (!syra_site.mobile) {
				var ii = 0,
					jj = $items.length;
				while (ii < jj) {
					var $item = $items[ii];
					if ($item && $item.$bind) {
						if ($colBinds.indexOf($item.$bind) < 0) {
							$items.splice(ii, 1);
							jj = $items.length;
							continue;
						} else {
							defined.splice(defined.indexOf($item.$bind), 1);
						}
					}
					ii++;
				}
				for (var ii = 0, jj = defined.length; ii < jj; ii++) {
					$items.push({
						$bind: defined[ii]
					});
				}
			}
		}
		if (this.list.$pagePreferences.$columns) {
			var source = this.list.$item.$layout.$items;
			this.list.$origColumns = [];
			for (var ii = 0, jj = $items.length; ii < jj; ii++) {
				this.list.$origColumns.push($items[ii].$bind);
			}

			var $items = this.list.$item.$layout.$items = [];
			for (var ii = 0, jj = this.list.$pagePreferences.$columns.length; ii < jj; ii++) {
				var $bind = this.list.$pagePreferences.$columns[ii];
				for (var mm = 0, kk = source.length; mm < kk; mm++) {
					var $item = source[mm];
					if ($item.$bind == $bind) {
						source.splice(mm, 1);
						$items.push($item);
						break;
					}
				}
			}
			for (var ii = 0, jj = source.length; ii < jj; ii++) {
				$items.push(source[ii]);
			}
		}

		_gridCardBuilder.load(this);
	},


	list_scrollToRecord: function(record) {
		if (record) {
			var view;
			if (this.list.$item && (this.list.$item.$isNavigationList || this.list.$item.$isSelBoxList)) {
				view = this.scrollTable.bodySlot;
			}
			record.page.scrollToItem(record.dataRow, view || record.page.scrollview, true);
		}

	},
	list_onAfterApplyDesignMetaData: function(metaData, designing) {
		if (designing && metaData.$cardPosition !== undefined) {
			_gridCardBuilder.applyDesignMeta(this, metaData, designing);
		}
		this.list.applyGrapDesignMetaData && this.list.applyGrapDesignMetaData(metaData, designing);
		this.list.paging.applyMetadata(metaData, designing);
	},
	applyMetaDataRecords: function($meta) {
		var $binds = Object.keys($meta.$properties);
		for (var ii = 0, jj = $binds.length; ii < jj; ii++) {
			var col = this.columnsMap[$binds[ii]];
			col && col.table.applyColumnMetaData(col, $meta.$properties[$binds[ii]]);
		}
		_checkHeaderVisibility(this);
	},
	list_onItemInOut: function(onEnter, event, target) {
		if (target.syra_field_bind) {
			var col = this.findEventCol(event);
			if (this.list.isColReordering) {
				if (syra_dd.ddAgent && syra_dd.ddAgent.onColumnHeaderEnterLeave) {
					syra_dd.ddAgent.onColumnHeaderEnterLeave(this.columnsMap[target.syra_field_bind], onEnter);
				}
			}
			if (onEnter) {
				if (col.titleCell.syraResizerColId) {
					col.rect = col.titleCell.getBoundingClientRect();
				}
			}
		}
		this.list.selector.onItemInOut(this, onEnter, event, target);
	},
	list_onMouseMove: function(event) {
		var col = this.findEventCol(event);
		col && _colResizer.setCursor(col, event);
	},
	isDraggable: function(target, event) {
		if (target.syraResizerColId == this.list.id) {
			if (_colResizer.isDraggable(this.list, this.findEventCol(event), event)) {
				return;
			}
		}
		if (target.syraReorderColId == this.list.id) {
			_colsReorderer.isDraggable(this.list, this.findEventCol(event), event);
			return;
		} else {
			if (this.list.addSlotResizer(target, event)) {
				return;
			} else {
				return this.list.isReorderRecordDraggable(target, event);
			}
		}
		return null;
	},
	record_remove: function(record) {
		this.fixedTable && this.fixedTable.removeRow(record);
		this.scrollTable && this.scrollTable.removeRow(record);
		this.record_disposeCard(record, true);
	},
	record_disposeCard: function(record, isRecordRemoved) {
		if (record.rowCard) {
			syra_dom.remove(record.rowCard.row);
			if (this.list.inlineCards) {
				var ii = this.list.inlineCards[record.rowCard];
				(ii >= 0) && this.list.inlineCards.splice(ii, 1);
			}
		}
		if (!isRecordRemoved) {
			syra_button.dispose(record.rowCardBtn);
			record.layoutSlot = record.layoutContent = record.domItem = record.body = record.rowCard = record.cardItem = null;
			delete record.cardItem;
		}
	},
	record_resizeInlineCard: function(card, width) {
		card.contentSlot.style.maxWidth = width;
	},
	record_draw: function(record) {
		record.isGridRecord = true;
		record.$mnPickersCss = record.$mnPickersCss || "s-list-record-picker";
		record.$iconPath = this.list.$iconPath;
		record.$isVerticalDirection = true;
		record.domItem = record.dataRow = _addRow(this, record);
		if (this.allColumns[0].table.isFixedCols) {
			record.fixedRow = _addRow(this, record);
		}
		record.fieldCellsMap = {};
		record.cellsMap = {};
		if (this.list.treeDecorator) {
			var col = this.columnsMap.$treeview;
			var td = record.cellsMap[col.key] = document.createElement("td");
			td.className = this.list.cssCell + " " + this.list.treeDecorator.cssNodeCell;
			if (this.list.$capability.reorder) {
				td.className += " s-tree-record-reorder";
				td.syraDragSpot = this.list.id;
			}
			record[col.table.$rowKey].appendChild(td);
			this.list.treeDecorator.addCardNode(record, td);
		} else {
			var before = record.insertBeforeRecord;
			if (before) {
				before.dataRow.parentNode.insertBefore(record.dataRow, before.dataRow);
				before.fixedRow && before.fixedRow.parentNode.insertBefore(record.fixedRow, before.fixedRow);
				delete record.insertBeforeRecord;
			} else {
				record.fixedRow && this.fixedTable.bodyTable.appendChild(record.fixedRow);
				this.scrollTable.bodyTable.appendChild(record.dataRow);
			}
		}

		var cssCell = this.list.cssCell;
		if (this.list.treeDecorator) {
			cssCell += " s-list-no-border";
		}
		for (var ii = 0, jj = this.allColumns.length; ii < jj; ii++) {
			var col = this.allColumns[ii];
			if (col == this.columnsMap.$rowIndex) {
				record.cellsMap[col.key] = _appendRowIndexCell(record, record[col.table.$rowKey]);
				continue;
			}
			if (col == this.columnsMap.rowCard) {
				record[col.table.$rowKey].appendChild(_gridCardBuilder.addRowCardOpener(record, col));
				continue;
			}
			if (col == this.columnsMap.$recordSelector) {
				var td = record.cellsMap[col.key] = document.createElement("td");
				td.className = this.list.cssCell;
				this.list.selector.appendToRecord(record, record[col.table.$rowKey].appendChild(td));
				continue;
			}
			if (col == this.columnsMap.$menus) {
				record.cellsMap[col.key] = record.menusSlot = document.createElement("td");
				record.menusSlot.className = this.list.cssCell + " " + this.list.$skin + "-actions-cell";
				record[col.table.$rowKey].appendChild(record.menusSlot);
				if (this.list.treeDecorator) {
					record.menusSlot.className += " s-list-tree-cell";
				}
				this.list.applyCapabilitiesToRecord(record);
				continue;
			}
			if (col.$bind) {
				var td = record.cellsMap[col.key] = document.createElement("td");
				td.className = cssCell;
				if (col.$isHidden) {
					syra_dom.hide(td, true);
				}
				record.fieldCellsMap[col.key] = record.page.addItem(td, col.$item, record);

				if (!this.list.$isEditMode && col.$field.$hasFilterCapability) {
					td.className += col.cssRecordCellFilter;
				}
				record[col.table.$rowKey].appendChild(td);
			}
		}

	},
	list_onAfterFill: function() {
		this.showOutCard && this.showOutCard();
	},

	list_onAfterDraw: function() {
		this.list.addSearcher();
		this.list.$skin = this.list.$item.$skin || this.list.$defaultSkin;
		this.list.cssCell = this.list.$skin + "-cell";
		if (this.list.$item.$cssCell) {
			this.list.cssCell += " " + this.list.$item.$cssCell;
		}
		if (this.list.$isEditMode) {
			this.list.cssCell += "-edit";
		}
		this.list.cssRowIndex = this.list.$skin + "-cell " + this.list.$skin + "-row-index";
		this.list._dataValue.style.overflow = "";
		this.tableSlot = syra_dom.addDiv("s-grid-slot", this.list._core);

		this.list.drawGraph();
		this.appendCardGraphSlots();


		this.list.applyDesignMeta(this.list.$item, false);
		this.list.setState(this.list.$field);

		this.slotTable = syra_dom.addDiv("s-grid-slot-table");

		this.scrollTable = _gridTable.add(this);
		this.slotTable.appendChild(this.scrollTable.slot);

		this.list.body = this.scrollTable.bodyTable;
		this.tableSlot.appendChild(this.slotTable);

		this.list.addSelector();
		this.list.applyCapabilities();
		this.list.ensureGlobalMetaRecord(true);
		_addTitleRow(this);

		this.list.cursor = "default";
		syra_dd.addToColResizers(this.list, true);
		this.emptyDataItem = syra_dom.addDiv(this.list.$skin + "-empty-slot");
		this.list.endDraw();
	},
	appendCardGraphSlots: function() {
		var $cardItem = this.list.$item.$format != "cards" && this.list.$item.$cardItem;
		!$cardItem && _gridCardBuilder.removeOutSlot(this);

		if (!this.list.$item.$graphPosition) {
			this.list.removeGrapSlot && this.list.removeGrapSlot();
		}
		if (!$cardItem && !this.list.$item.$graphPosition) {
			if (this.list.tableRow) {
				syra_dom.moveChildNodes(this.list.tableRow, this.list._core);
				this.list._core.removeChild(this.list.tableRow);
				delete this.list.tableRow;
			}
		}
		if ($cardItem || this.list.$item.$graphPosition) {
			this.list.fitWidth = true;
			var hasRow = $cardItem && ($cardItem.$position == "left" || $cardItem.$position == "right");
			if (!hasRow) {
				hasRow = this.list.$item.$graphPosition && (this.list.$item.$graphPosition == "left" || this.list.$item.$graphPosition == "right");
			}
			if (hasRow) {
				if (!this.list.tableRow) {
					this.list.tableRow = syra_dom.addDiv("s-grid-table-row");
					syra_dom.moveChildNodes(this.list._core, this.list.tableRow);
					this.list._core.appendChild(this.list.tableRow);
				}
			} else {
				if (this.list.tableRow) {
					syra_dom.moveChildNodes(this.list.tableRow, this.list._core);
					this.list._core.removeChild(this.list.tableRow);
					delete this.list.tableRow;
				}
			}
			$cardItem && _gridCardBuilder.addOutSlot(this);

			if (this.list.$item.$graphPosition) {
				this.list.addGrapSlot && this.list.addGrapSlot();
			}
		}
		if (this.list.tableRow) {
			this.tableSlot.className = "s-grid-slot-cell";
		} else {
			this.tableSlot.style.width = "";
			this.tableSlot.className = "s-grid-slot";
		}
	},
	getNextVisibleField: function($bind) {
		return this.getNextField($bind, true);
	},
	getNextFocusableField: function($bind) {
		return this.getNextField($bind, true, true);
	},
	getPreviousVisibleField: function($bind) {
		return this.getPreviousField($bind, true);
	},
	getPreviousFocusableField: function($bind) {
		return this.getPreviousField($bind, true, true);
	},
	getNextField: function($bind, visible, focusable) {
		var col = $bind && this.columnsMap[$bind];
		for (var ii = col ? this.allColumns.indexOf(col) + 1 : 0, jj = this.allColumns.length; ii < jj; ii++) {
			col = this.allColumns[ii];
			if (col.$bind) {
				if (!(col.$isHidden && (visible || focusable)) && !(col.$isDisabled && focusable) && (col.$field && col.$field.$type && col.$field.$type !== "image" && visible && focusable)) {
					break;
				}
			}
			col = null;
		}
		return (col && col.$bind) || $bind;
	},
	getPreviousField: function($bind, visible, focusable) {
		var col = $bind && this.columnsMap[$bind];
		for (var ii = col ? this.allColumns.indexOf(col) - 1 : this.allColumns.length - 1; ii >= 0; ii--) {
			col = this.allColumns[ii];
			if (col.$bind) {
				if ((col.$isHidden && (visible || focusable)) || (col.$isDisabled && focusable)) {
					col = null;
					continue;
				}
				break;
			}
		}
		return (col && col.$bind) || $bind;
	},

	record_highlightOnEnter: function(record, onEnter, event) {
		if (this.list.selector) {
			if (record.treeNode) {
				syra_dom.toggleClass(record.treeNode.desc || record.treeNode.item, "s-record-enter", onEnter);
				for (var ii = 0, jj = this.allColumns.length; ii < jj; ii++) {
					var col = this.allColumns[ii];
					if (!col.$isHidden && !col.isTreeview && col.key) {
						syra_dom.toggleClass(record.cellsMap[col.key], "s-record-enter", onEnter);
					}
				}
			} else {
				record.fixedRow && syra_dom.toggleClass(record.fixedRow, "s-record-enter", onEnter);
				syra_dom.toggleClass(record.dataRow, "s-record-enter", onEnter);
			}
			this.list.selector.onItemInOut(record, onEnter, event);
		}
	},
	record_onSelect: function(record, selected) {
		if (this.list.treeDecorator) {
			syra_dom.toggleClass(record.treeNode.desc || record.treeNode.item, "s-list-record-selected", selected);
			for (var ii = 0, jj = this.allColumns.length; ii < jj; ii++) {
				var col = this.allColumns[ii];
				if (!col.$isHidden && !col.isTreeview && col.key) {
					syra_dom.toggleClass(record.cellsMap[col.key], "s-list-record-selected", selected);
				}
			}
		} else {
			syra_dom.toggleClass(record.dataRow, "s-list-record-selected", selected);
		}
		record.fixedRow && syra_dom.toggleClass(record.fixedRow, "s-list-record-selected", selected);
		record.checkBox && syra_fields.checkbox.setValue(record, selected);
		if (selected && !this.list.selector.isMulti) {
			this.showOutCard && this.showOutCard(record);
			record.scrollToRecord();
		}
	},
	dispose: function(removeItem) {
		if (removeItem) {
			syra_dom.remove(this.tableSlot);
			_gridCardBuilder.removeOutSlot(this);
		}
		syra_dd.addToColResizers(this.list, false);
		if (this.list) {
			syra_item.remove(this.filterRecord);
			_gridFilter.dispose();
			_popupCardBuilder.dispose(this.list);
			this.list.quickEdit && this.list.quickEdit.dispose();
		}
		this.fixedTable && this.fixedTable.dispose();
		if (this.allColumns) {
			for (var ii = 0, jj = this.allColumns.length; ii < jj; ii++) {
				syra_site.disposeObject(this.allColumns[ii]);
			}
		}
		this.scrollTable && this.scrollTable.dispose();
		syra_site.disposeObject(this);
	},
	adjustChartSize: function(maxHeight) {
		if (this.list.chartField) {
			switch (this.list.$item.$graphPosition) {
				case "bottom":
				case "top":
					maxHeight /= 2;
					this.list.chartField.layoutSlot.style.height = maxHeight + "px";
					break;
			}
		}
		return maxHeight;
	},
	onResizeArray: function() {
		if (syra_dom.isVisible(this.tableSlot)) {
			var container = this.tableSlot;
			var scrollBodySlot = this.scrollTable.bodySlot;
			var scrollview, adjustToScrollview = this.list.adjustToScrollPage;
			if (this.list.isMaximized) {
				scrollview = this.list.domItem.parentNode;
				adjustToScrollview = true;
			} else {
				scrollview = this.list.page.scrollview;
				if (this.list.page.vignetteField) {
					if (!this.list.page.vignetteField.isMaximized) {
						adjustToScrollview = false;
						scrollview = this.list.page.vignetteField.page.scrollview;
					}
				}
			}
			if (adjustToScrollview && this.list.chartField) {
				if (this.list.$item.$graphPosition == "bottom" || this.list.$item.$graphPosition == "top") {
					adjustToScrollview = false;
				}
			}
			if (!this.list.tableRow) {
				container = this.tableSlot.parentNode; //if row tableSlott is cell so maxWidth
			}
			if (scrollview) {
				var pageSize = this.list.page.getPageSize();
				var maxSlotWidth = container.clientWidth; // container.getBoundingClientRect().width;
				var maxHeight;
				this.tableSlot.style.maxWidth = "";
				if (this.list.$item.$fixedBodyHeight) {
					var height = scrollBodySlot.clientHeight;
					scrollBodySlot.style.height = ((Math.abs(parseInt(this.list.$item.$fixedBodyHeight, 10) - height) < 2) ? height : this.list.$fixedBodyHeight) + "px";
				} else {
					if (adjustToScrollview) {
						scrollBodySlot.style.maxHeight = "";
						var scrollH = scrollview.scrollHeight;
						var diff = scrollH - scrollview.clientHeight;
						if (diff > 1) {
							scrollBodySlot.style.maxHeight = maxHeight = Math.max((scrollBodySlot.clientHeight - diff) - 2, 50) + "px";
						}
					} else {
						if (this.list.$fitContainer) {
							if (!this.emptyDataItem.parentNode) {
								var top = scrollBodySlot.scrollTop;
								var bodyRect = scrollBodySlot.getBoundingClientRect();
								var slotRect = this.list.layoutSlot.getBoundingClientRect();
								var height = Math.max(slotRect.height - (bodyRect.top - slotRect.top), 0);
								scrollBodySlot.style.height = height + "px";
								scrollBodySlot.scrollTop = top;
							}
						} else {
							if (scrollview) {
								var maxH;
								if (this.list.page.isLookUpPage) {
									maxH = pageSize.overPanel.maxHeight;
									maxH -= (this.scrollTable.headSlot.clientHeight + this.list._core.offsetTop + pageSize.lookupMargin);
								} else {
									maxH = pageSize.maxScrollHeight || scrollview.clientHeight;
									maxH -= (this.scrollTable.headSlot.clientHeight + 100);
								}
								scrollBodySlot.style.maxHeight = maxHeight = Math.max(maxH, 100) + "px";
							}
						}
					}
				}
				if (this.emptyDataItem.parentNode && (this.list.treeDecorator || this.list.$item.$isTitleRowHidden)) {
					this.tableSlot.style.maxWidth = this.emptyDataItem.clientWidth + "px";
				} else {
					var fixedWidth = 0;
					if (this.fixedTable) {
						var table = this.list.$item.$isTitleRowHidden ? this.fixedTable.bodyTable : this.fixedTable.headTable;
						this.fixedTable.setColWidths();
						fixedWidth = table.clientWidth + 1;
						this.fixedTable.slot.style.width = fixedWidth + "px";
						var hasMoved;
						while (this.fixedTable.columns.length && Math.max(maxSlotWidth - fixedWidth, 0) <= 300) {
							hasMoved = true;
							var col = this.fixedTable.columns.pop();
							fixedWidth -= col.colWidth;
							_colsReorderer.moveToSrollTable(col);
						}
						if (hasMoved) {
							if (this.fixedTable.columns.length) {
								fixedWidth = table.clientWidth + 1;
								this.fixedTable.slot.style.width = fixedWidth + "px";
							} else {
								fixedWidth = 0;
								syra_dom.remove(this.fixedTable.slot);
								syra_site.disposeObject(this.fixedTable);
								delete this.fixedTable;
							}
						}
					}
					var fitWidth;
					if (this.list.$isEditMode) {
						var definedWidth;
						for (var ii = 0, jj = this.scrollTable.columns.length; !definedWidth && ii < jj; ii++) {
							var col = this.scrollTable.columns[ii];
							if (col.$field && !col.$isHidden) {
								definedWidth = col.$field.$displayLength || col.$field.$maxLength || 0;
							}
						}
						fitWidth = !definedWidth;
					}
					var hasScollHeight = this.scrollTable.bodySlot.scrollHeight > (this.scrollTable.bodySlot.clientHeight + 1);
					var scrollLeft = scrollBodySlot.scrollLeft;
					if (!scrollLeft) {
						scrollLeft = this.scrollTable.headSlot.scrollLeft;
					}
					if (!fitWidth) {
						fitWidth = this.list.fitWidth || this.list.isMaximized;
					}
					if (!fitWidth) {
						fitWidth = this.list.$field.$isFusionNavigationList && !this.list.page.isLookUpPage;
					}
					if (fitWidth) {
						fitWidth = maxSlotWidth - fixedWidth - 1; //1 is margin security
						if (hasScollHeight) {
							fitWidth -= 17;
						}
						this.scrollTable.setColWidths(fitWidth && fitWidth);
					} else {
						this.scrollTable.setColWidths();
					}

					if (hasScollHeight) {
						if (!this.scrollTable._headerCorner) {
							this.scrollTable._headerCorner = document.createElement("div");
							this.scrollTable._headerCorner.className = this.scrollTable.builder.list.$skin + "-head-corner";
							this.scrollTable.headSlot.parentNode.appendChild(this.scrollTable._headerCorner);
						}
						this.scrollTable._headerCorner.style.height = this.scrollTable.headSlot.clientHeight + "px";
					}
					syra_dom.hide(this.scrollTable._headerCorner, !hasScollHeight);

					this.scrollTable.headSlot.style[syra_site.isRTL ? "marginLeft" : "marginRight"] = hasScollHeight ? "17px" : "";
					scrollBodySlot.style[syra_site.isRTL ? "paddingLeft" : "paddingRight"] = hasScollHeight ? "17px" : "";

					var scrollWidth = this.scrollTable.bodyTable.clientWidth;
					if (!scrollWidth) {
						scrollWidth = this.scrollTable.headTable.clientWidth;
					}
					scrollWidth += ((this.scrollTable._headerCorner && 17) || 0);

					scrollWidth += (fixedWidth || 0);


					if (scrollWidth <= 1) {
						scrollWidth = "";
					}
					scrollBodySlot.scrollLeft = scrollLeft;
					this.tableSlot.style.maxWidth = scrollWidth + "px";
					if (!this.list.tableRow) {
						this.list.topbar.style.maxWidth = scrollWidth + "px";
					}
					if (scrollWidth && !this._slotTableWidh) {
						this.slotTable.style.width = this._slotTableWidh = "100%";
					}
					var fixedBody = this.fixedTable && this.fixedTable.bodySlot;
					if (fixedBody) {
						var height = scrollBodySlot.clientHeight;
						if (!height && this.list.records.length) {
							fixedBody.style.maxHeight = ""; //manage case where there is no scroll column visible but only fixed column 
						} else {
							fixedBody.style.maxHeight = height + "px";
						}
					}
					if (this.emptyDataItem.parentNode) {
						this.emptyDataItem.style.width = (this.scrollTable.headTable.scrollWidth - 1) + "px";
					}
				}

			}
			if (this.list.inlineCards) {
				var width = this.scrollTable.bodySlot.clientWidth;
				if (this.scrollTable.bodySlot.scrollWidth != width) {
					width -= 17;
				}
				width = (width - (16 + 4)) + "px";
				for (var ii = 0, jj = this.list.inlineCards.length; ii < jj; ii++) {
					this.record_resizeInlineCard(this.list.inlineCards[ii], width);
				}
			}
		}
	},
	clearPreferences: function() {
		_colsReorderer.resetReorder(this.list);
		delete this.list.$pagePreferences.$columnWidths;
		if (this.list.page.isVignettePage) {
			for (var ii = 0, jj = this.allColumns.length; ii < jj; ii++) {
				if (this.allColumns[ii].$item) {
					delete this.allColumns[ii].$item.$width;
				}
			}
		}
	},
	toggleColWidthToContent: function() {

	},
	record_reorder: function(record, target, isAfter) {
		this.fixedTable && this.fixedTable.reorderRow(target, record, isAfter);
		this.scrollTable && this.scrollTable.reorderRow(target, record, isAfter);
		record.rowCard && record.dataRow.parentNode.insertBefore(record.rowCard.row, record.dataRow.nextSibling);
		if (this.list.quickEdit && this.list.quickEdit.record == record) {
			record.dataRow.parentNode.insertBefore(this.list.quickEdit.card.row, record.dataRow.nextSibling);
		}
	},
	onChildFieldFocus: function(field, isFocusIn) {
		var record = field.articleParent,
			self = this;
		if (record.rowIndexBtn && syra_site.isTabletDevice) {
			//for classicPage, esc is done by default on click
			// but escape is remove on field focusout it's too early so setTimeout ! 
			//quite ugly but if somebody find a best way without breaking the code, let's go !
			setTimeout(function() {
				self.tooggleEscape(record, isFocusIn);
			}, 200);
		}
	},
	tooggleEscape: function(record, isFocusIn) {
		syra_dom.toggleClass(record.rowIndexBtn, "s-grid-row-escape", record._hasEscapeRowButton = isFocusIn);
		record.setRowIndex();
	}
});
