// syracuse-ui/lib/field/array/listContextMenu.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function ListContextMenu(){
}

exports.ListContextMenu = helpers.defineClass(ListContextMenu, ContextMenu, {

    create: function(field){
        var self = this
        if (field.list) {
            self.hasColumns = false;
            switch (field.list.$item.$format) {
                case "vgrid":
                    self.popupContent = {
                        id: helpers.uuid.generate(),
                        $$item: $(field._menusBoxSlot)
                    };
                    break;
                case "grid":
                    self.hasColumns = true;
                    break;
            }
        }
        
        ContextMenu.prototype.create.call(this, field, {
            onBeforeAddMenuItem: function($menuItem){
            
                if (!self._isMenuBtnVisible) {
                
                    self._isMenuBtnVisible = true;
                    
                    if ($menuItem.$bind == "$lazyload") {
                        var title;
                        if (self.field.list) {
                            title = self.field.list._localize.mn_actions;
                        }
                        else {
                            title = self.field._localize.mn_actions;
                            self.menuPicker.textContent = title;
                        }
                        self.menuPicker.setAttribute("title", title);
                        return false;
                    }
                }
                
            }
        });
    },
    appendOpener: function(){
        this.menuPicker = document.createElement("a");
        var title;
        if (this.field.list) {
            this.menuPicker.className = "s-record-menus-btn s-close";
            title = this.field.list._localize.mn_more;
        }
        else {
            this.menuPicker.className = "s-list-menus-btn s-close";
            title = this.field._localize.mn_more;
            this.menuPicker.textContent = title;
        }
        this.menuPicker.setAttribute("title", title);
        this.menuPicker.setAttribute("data-s-picker", "menus");
        this._isMenuBtnVisible = false;
        this.menuPicker.setAttribute("id", this.field.id + "-menu-btn");
        (this._menusBtnSlot = document.createElement("div")).className = "s-quick-actions-slot";
        this._menusBtnSlot.style.display = "none";
        this._actionsSlot.appendChild(this._menusBtnSlot).appendChild(this.menuPicker);
    }
});
