// syracuse-ui/lib/field/array/recordArticle.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Article = require("syracuse-ui/lib/article/article").Article;
var ListContextMenu = require('./listContextMenu').ListContextMenu;

function RecordArticle(){
}

exports.RecordArticle = helpers.defineClass(RecordArticle, Article, {
    notifyChange: function($bind, value,isNotifyOnly){
        this.list._store.notifyRecordChange(this, $bind, value,isNotifyOnly);
    },
    applyActionLinkChange: function(resources, record){
        Article.prototype.applyActionLinkChange.call(this, resources, record);
        if (resources && (resources.$links !== undefined || resources.$actions !== undefined)) {
            if (!this._contextMenu && !this.$item.$isMenuRecordHidden) {
                (this._contextMenu = new ListContextMenu()).create(this);
            }
            if (resources.$links === null && this._contextMenu) {
                this._contextMenu.clearMenuItems();
            }
        }
    },
    renderRowIndex: function(){
        if (this._rowIndexCell) {
            this._rowIndexCell.textContent = (this.list._store.pagingCapability._$startIndex || 0) + this.$recordIndex;
        }
    },
    applyMetaData: function(metaData){
    },
    dispose: function(){
        delete this.list
        delete this.builder;
        if (this._contextMenu) {
            document.controller.disposeObject(this._contextMenu);
            delete this._contextMenu;
        }
        Article.prototype.dispose.call(this);
    }
});
