// syracuse-ui/lib/field/array/singleArrayField.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var ArrayField = require("syracuse-ui/lib/field/array/arrayField").ArrayField;
var SingleArrayBuilder = require("syracuse-ui/lib/field/array/singleArrayBuilder").SingleArrayBuilder;

function validatetemp(list, newValue, errors) {
	if (!list.$prototype.$isMultiValue && (list.$prototype.$item.$type != "application/x-boolean")) {
		var values = [];
		for (var ii = 0, jj = list.records.length; ii < jj; ii++) {
			//values = 
		}
		var newValue = field.getValue();
		for (var ii = 0, jj = list.records.length; ii < jj; ii++) {
			var record = list.records[ii];
			if (record != list) {
				var valueComp = record.singleField.getValue();
				// in case of reference field
				if (record.singleField.isReferenceField) {
					if (newValue[record.singleField.$reference.$value.$prop] == valueComp[record.singleField.$reference.$value.$prop]) {
						errors.push(syra_local.sglRecordUniqueValue);
					}
				} else {
					if (valueComp == newValue) {
						errors.push(syra_local.sglRecordUniqueValue);
					}
				}
			}
		}
	}
}


function SingleArrayField() {}

exports.SingleArrayField = helpers.defineClass(SingleArrayField, ArrayField, {
	loadBox: function(initData) {
		if (!this.$field.$builderClass) {
			this.$field.$builderClass = SingleArrayBuilder;
		}
		ArrayField.prototype.loadBox.call(this, initData);
	},
	initializeList: function() {
		this.formType = "singleArray";
		this.isPagingDisabled = true;
		this.isClientPagingPerPageDisable = true;
		this.isSingList = true;
		this.isServerStepPager = false;
		this.isClientFetch = this.isClientFetch !== false;
		syra_fields.ensureEditMode(this);
	},
	emptyBody: function() {
		syra_dom.remove(this.topbar);
		this.filler.emptyBody(this);
		this.addMenusBar();
	},
	fillList: function(dataRecordSet, parentDataRecord, isDelta) {
		var focusRecordIndex = -1;
		if (this.page.focusField && this.page.focusField.articleParent &&
			this.page.focusField.articleParent.list == this) {
			focusRecordIndex = this.page.focusField.articleParent.getRecordIndex();
		}
		this.paging.applyMetadata(parentDataRecord);
		syra_dom.remove(this.topbar);

		this.filler.fill(this, dataRecordSet, parentDataRecord, isDelta);
		this.addMenusBar();
		this.showBody(true);
		if (focusRecordIndex >= 0) {
			if (this.records.length) {
				this.records[Math.min(focusRecordIndex, this.records.length - 1)].boundFields.$singleField[0].focus();
			}
		}
	},
	endDraw: function() {
		syra_dd.addToColResizers(this, true);
		this.layoutSlot.appendChild(this.domItem);
		this.$skin = this.$item.$skin || "s-single";
		this.body = syra_dom.addDiv(this.$skin + "-body");
		syra_dom.remove(this.topbarRightSlot); // not used for single field
		this.topbar.className = this.$skin + "-menus";
		this._core.appendChild(this.body);
		this.applyDesignMeta(this.$item, false);
		this.setState(this.$field);
		this.applyCapabilities();
		this.ensureGlobalMetaRecord(true);
		syra_fields.ensureLayoutMode(this);
	},
	addMenusBar: function() {
		if (this.$isMenuBarAtTop) {
			if (this.bodyInsert) {
				this.bodyInsert.parentNode.insertBefore(this.topbar, this.bodyInsert.parentNode.firstChild);
			} else {
				this.body.parentNode.insertBefore(this.topbar, this.body);
			}
		} else {
			this.body.appendChild(this.topbar);
		}
	},
	dispose: function() {
		syra_dd.addToColResizers(this, false);
		ArrayField.prototype.dispose.call(this);
	}
});
