// @sage/syracuse-react/src/test/navigation/sitemap/sitemapTest.tsx
"use strict";
import * as React from 'react';
import Section from '../../../section/section';
import { Sitemap } from '../../../navigation/sitemap/sitemap';
import { SitemapModule, SitemapSubmodule } from '../../../navigation/sitemap/groups';
import { SitemapMenu, SitemapBlock } from '../../../navigation/sitemap/items';
import { prototype } from './prototype';
import { dataset } from './dataset';
import SitemapPage from '../../../navigation/sitemap/page/sitemapPage';

export default class SitemapTest extends React.Component<{}, {}>{
	static representationName: string = "sitemap";
	searchResult: any;
	onSelectModule = (module: SitemapModule) => {

	}
	onSwitchBookmark = (menu: SitemapMenu) => {
	}
	onMenuClick = (menu: SitemapMenu, event: MouseEvent) => {

	}
	onModuleCollapse = (module: SitemapModule) => {
		module.collapsed = !module.collapsed;
		this.setState({});
	}
	onSubModuleCollapse = (submodule: SitemapSubmodule) => {
		submodule.collapsed = !submodule.collapsed;
		this.setState({});
	}
	onMenuBlockCollapse = (menuBlock: SitemapBlock) => {
		menuBlock.collapsed = !menuBlock.collapsed;
		this.setState({});
	}
	render() {
		let sitemap = new Sitemap(prototype, dataset);
		//let modules: SitemapModule[] = sitemap.modules;
		let modules: any = [{
			$uuid: "1",
			title: "Module 1",
			selected: true,
			submodules: [{
				$uuid: "11",
				title: "Module 1 - Submodule 1",
				children: [
				]
			}, {
				$uuid: "12",
				title: "Module 1 - Submodule 2",
				children: []
			}]
		}];
		
		return <SitemapPage
			userFirstName={"User Test"}
			modules={modules}
			onModuleBodyScroll={() => { }}
			onModuleCollapse={this.onModuleCollapse}
			onSubModuleCollapse={this.onSubModuleCollapse}
			onMenuBlockCollapse={this.onMenuBlockCollapse}
			onSelectModule={this.onSelectModule}
			onMenuClick={this.onMenuClick}
			onSwitchBookmark={this.onSwitchBookmark}
			onCollapseAllModules={() => { }}
			onSearch = {() =>{}}
			searchResult={this.searchResult}
		/>;
	}
}
