// hdb/lib-cov/protocol/auth/SAML.js
/* automatically generated by JSCoverage - do not edit */
if (typeof _$jscoverage === 'undefined') _$jscoverage = {};
if (! _$jscoverage['protocol/auth/SAML.js']) {
  _$jscoverage['protocol/auth/SAML.js'] = [];
  _$jscoverage['protocol/auth/SAML.js'][14] = 0;
  _$jscoverage['protocol/auth/SAML.js'][16] = 0;
  _$jscoverage['protocol/auth/SAML.js'][18] = 0;
  _$jscoverage['protocol/auth/SAML.js'][19] = 0;
  _$jscoverage['protocol/auth/SAML.js'][20] = 0;
  _$jscoverage['protocol/auth/SAML.js'][21] = 0;
  _$jscoverage['protocol/auth/SAML.js'][22] = 0;
  _$jscoverage['protocol/auth/SAML.js'][23] = 0;
  _$jscoverage['protocol/auth/SAML.js'][26] = 0;
  _$jscoverage['protocol/auth/SAML.js'][27] = 0;
  _$jscoverage['protocol/auth/SAML.js'][30] = 0;
  _$jscoverage['protocol/auth/SAML.js'][31] = 0;
  _$jscoverage['protocol/auth/SAML.js'][34] = 0;
  _$jscoverage['protocol/auth/SAML.js'][35] = 0;
  _$jscoverage['protocol/auth/SAML.js'][38] = 0;
  _$jscoverage['protocol/auth/SAML.js'][39] = 0;
}
_$jscoverage['protocol/auth/SAML.js'][14]++;
"use strict";
_$jscoverage['protocol/auth/SAML.js'][16]++;
module.exports = SAML;
_$jscoverage['protocol/auth/SAML.js'][18]++;
function SAML(options) {
  _$jscoverage['protocol/auth/SAML.js'][19]++;
  options = options || {};
  _$jscoverage['protocol/auth/SAML.js'][20]++;
  this.name = "SAML";
  _$jscoverage['protocol/auth/SAML.js'][21]++;
  this.assertion = options.assertion || options.password;
  _$jscoverage['protocol/auth/SAML.js'][22]++;
  this.user = undefined;
  _$jscoverage['protocol/auth/SAML.js'][23]++;
  this.sessionCookie = undefined;
}
_$jscoverage['protocol/auth/SAML.js'][26]++;
SAML.prototype.initialData = (function initialData() {
  _$jscoverage['protocol/auth/SAML.js'][27]++;
  return this.assertion;
});
_$jscoverage['protocol/auth/SAML.js'][30]++;
SAML.prototype.initialize = (function initialize(buffer) {
  _$jscoverage['protocol/auth/SAML.js'][31]++;
  this.user = buffer.toString("utf8");
});
_$jscoverage['protocol/auth/SAML.js'][34]++;
SAML.prototype.finalData = (function finalData() {
  _$jscoverage['protocol/auth/SAML.js'][35]++;
  return new Buffer(0);
});
_$jscoverage['protocol/auth/SAML.js'][38]++;
SAML.prototype.finalize = (function finalize(buffer) {
  _$jscoverage['protocol/auth/SAML.js'][39]++;
  this.sessionCookie = buffer;
});
_$jscoverage['protocol/auth/SAML.js'].source = ["// Copyright 2013 SAP AG.","//","// Licensed under the Apache License, Version 2.0 (the \"License\");","// you may not use this file except in compliance with the License.","// You may obtain a copy of the License at","//","// http: //www.apache.org/licenses/LICENSE-2.0","//","// Unless required by applicable law or agreed to in writing,","// software distributed under the License is distributed on an","// \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,","// either express or implied. See the License for the specific","// language governing permissions and limitations under the License.","'use strict';","","module.exports = SAML;","","function SAML(options) {","  options = options || {};","  this.name = 'SAML';","  this.assertion = options.assertion || options.password;","  this.user = undefined;","  this.sessionCookie = undefined;","}","","SAML.prototype.initialData = function initialData() {","  return this.assertion;","};","","SAML.prototype.initialize = function initialize(buffer) {","  this.user = buffer.toString('utf8');","};","","SAML.prototype.finalData = function finalData() {","  return new Buffer(0);","};","","SAML.prototype.finalize = function finalize(buffer) {","  this.sessionCookie = buffer;","};"];
