// hdb/lib-cov/protocol/auth/SessionCookie.js
/* automatically generated by JSCoverage - do not edit */
if (typeof _$jscoverage === 'undefined') _$jscoverage = {};
if (! _$jscoverage['protocol/auth/SessionCookie.js']) {
  _$jscoverage['protocol/auth/SessionCookie.js'] = [];
  _$jscoverage['protocol/auth/SessionCookie.js'][14] = 0;
  _$jscoverage['protocol/auth/SessionCookie.js'][16] = 0;
  _$jscoverage['protocol/auth/SessionCookie.js'][18] = 0;
  _$jscoverage['protocol/auth/SessionCookie.js'][20] = 0;
  _$jscoverage['protocol/auth/SessionCookie.js'][21] = 0;
  _$jscoverage['protocol/auth/SessionCookie.js'][22] = 0;
  _$jscoverage['protocol/auth/SessionCookie.js'][23] = 0;
  _$jscoverage['protocol/auth/SessionCookie.js'][24] = 0;
  _$jscoverage['protocol/auth/SessionCookie.js'][25] = 0;
  _$jscoverage['protocol/auth/SessionCookie.js'][26] = 0;
  _$jscoverage['protocol/auth/SessionCookie.js'][27] = 0;
  _$jscoverage['protocol/auth/SessionCookie.js'][30] = 0;
  _$jscoverage['protocol/auth/SessionCookie.js'][31] = 0;
  _$jscoverage['protocol/auth/SessionCookie.js'][34] = 0;
  _$jscoverage['protocol/auth/SessionCookie.js'][38] = 0;
  _$jscoverage['protocol/auth/SessionCookie.js'][39] = 0;
  _$jscoverage['protocol/auth/SessionCookie.js'][42] = 0;
}
_$jscoverage['protocol/auth/SessionCookie.js'][14]++;
"use strict";
_$jscoverage['protocol/auth/SessionCookie.js'][16]++;
var os = require("os");
_$jscoverage['protocol/auth/SessionCookie.js'][18]++;
module.exports = SessionCookie;
_$jscoverage['protocol/auth/SessionCookie.js'][20]++;
function SessionCookie(options) {
  _$jscoverage['protocol/auth/SessionCookie.js'][21]++;
  options = options || {};
  _$jscoverage['protocol/auth/SessionCookie.js'][22]++;
  this.name = "SessionCookie";
  _$jscoverage['protocol/auth/SessionCookie.js'][23]++;
  var pid = process.pid || "nodejs";
  _$jscoverage['protocol/auth/SessionCookie.js'][24]++;
  var termId = new Buffer(pid + "@" + os.hostname(), "utf8");
  _$jscoverage['protocol/auth/SessionCookie.js'][25]++;
  var rawSessionCookie = options.sessionCookie;
  _$jscoverage['protocol/auth/SessionCookie.js'][26]++;
  var length = rawSessionCookie.length + termId.length;
  _$jscoverage['protocol/auth/SessionCookie.js'][27]++;
  this.sessionCookie = new Buffer.concat([rawSessionCookie, termId], length);
}
_$jscoverage['protocol/auth/SessionCookie.js'][30]++;
SessionCookie.prototype.initialData = (function initialData() {
  _$jscoverage['protocol/auth/SessionCookie.js'][31]++;
  return this.sessionCookie;
});
_$jscoverage['protocol/auth/SessionCookie.js'][34]++;
SessionCookie.prototype.initialize = (function initialize(buffer) {
});
_$jscoverage['protocol/auth/SessionCookie.js'][38]++;
SessionCookie.prototype.finalData = (function finalData() {
  _$jscoverage['protocol/auth/SessionCookie.js'][39]++;
  return new Buffer(0);
});
_$jscoverage['protocol/auth/SessionCookie.js'][42]++;
SessionCookie.prototype.finalize = (function finalize(buffer) {
});
_$jscoverage['protocol/auth/SessionCookie.js'].source = ["// Copyright 2013 SAP AG.","//","// Licensed under the Apache License, Version 2.0 (the \"License\");","// you may not use this file except in compliance with the License.","// You may obtain a copy of the License at","//","// http: //www.apache.org/licenses/LICENSE-2.0","//","// Unless required by applicable law or agreed to in writing,","// software distributed under the License is distributed on an","// \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,","// either express or implied. See the License for the specific","// language governing permissions and limitations under the License.","'use strict';","","var os = require('os');","","module.exports = SessionCookie;","","function SessionCookie(options) {","  options = options || {};","  this.name = 'SessionCookie';","  var pid = process.pid || 'nodejs';","  var termId = new Buffer(pid + '@' + os.hostname(), 'utf8');","  var rawSessionCookie = options.sessionCookie;","  var length = rawSessionCookie.length + termId.length;","  this.sessionCookie = new Buffer.concat([rawSessionCookie, termId], length);","}","","SessionCookie.prototype.initialData = function initialData() {","  return this.sessionCookie;","};","","SessionCookie.prototype.initialize = function initialize(buffer) {","  /* jshint unused:false */","};","","SessionCookie.prototype.finalData = function finalData() {","  return new Buffer(0);","};","","SessionCookie.prototype.finalize = function finalize(buffer) {","  /* jshint unused:false */","};"];
