// msoffice/lib/ppt/html/pptInterfaceImpl.js
"use strict";
var helpers = require('syracuse-core').helpers;
var base64 = require("msoffice/lib/Base64").Base64;

function PptInterfaceImpl() {}

exports.PptInterfaceImpl = helpers.defineClass(PptInterfaceImpl, null, {
	getResourceUrl: function() {
		return external.getPptCustomXlsData().getResourceUrl();
	},
	addDataToWorksheet: function(data) {
		var jsonData = JSON.stringify(data);
		external.addDataToWorksheet(jsonData);
	},
	getDocumentTitle: function() {
		return external.getPptCustomData().getDocumentTitle();
	},
	setDocumentTitle: function(value) {
		external.getPptCustomData().setDocumentTitle(value);
		external.getPptCustomData().writeDictionaryToDocument();
	},
	getDocumentUrl: function() {
		return external.getPptCustomData().getDocumentUrl();
	},
	setDocumentUrl: function(value) {
		external.getPptCustomData().setDocumentUrl(value);
		external.getPptCustomData().writeDictionaryToDocument();
	},
	refreshNextChart: function() {
		external.refreshNextChart();
	},
	getActionType: function() {
		var actionType = external.getPptCustomData().getActionType();
		return actionType;
	},
	expectedVersion: function(value) {
		if ("expectedVersion" in external) {
			external.expectedVersion(value);
		}
	},
	getAddinVersion: function() {
		return external.getAddinVersion();
	}
});
