// syracuse-ui/lib/common/article/layouts/compositeLayout.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Layout = require('./layout').Layout;
var layoutHelper = require('./layoutHelper');

var _$layouts = {
    "composite-3-1": {
        $layoutType: "stack",
        $items: [{
            $layoutType: "columns",
            $layoutSubType: "50,50",
            $items: [{
                $layoutType: "stack",
                $items: []
            }, {
                $layoutType: "stack",
                $items: []
            }]
        }, {
            $layoutType: "tabs",
            $items: []
        }, {
            $layoutType: "columns",
            $layoutSubType: "50,50",
            $items: [{
                $layoutType: "stack",
                $items: []
            }, {
                $layoutType: "stack",
                $items: []
            }]
        }]
    },
    "composite-3-2": {
        $layoutType: "stack",
        $items: [{
            $layoutType: "columns",
            $layoutSubType: "50,50",
            $items: [{
                $layoutType: "stack",
                $items: []
            }, {
                $layoutType: "stack",
                $items: []
            }]
        }, {
            $layoutType: "columns",
            $layoutSubType: "50,50",
            $items: [{
                $layoutType: "stack",
                $items: []
            }, {
                $layoutType: "stack",
                $items: []
            }]
        }, {
            $layoutType: "columns",
            $layoutSubType: "50,50",
            $items: [{
                $layoutType: "stack",
                $items: []
            }, {
                $layoutType: "stack",
                $items: []
            }]
        }]
    },
    "composite-4-1": {
        $layoutType: "stack",
        $items: [{
            $layoutType: "columns",
            $layoutSubType: "50,50",
            $items: [{
                $layoutType: "stack",
                $items: []
            }, {
                $layoutType: "stack",
                $items: []
            }]
        }, {
            $layoutType: "columns",
            $layoutSubType: "50,50",
            $items: [{
                $layoutType: "stack",
                $items: []
            }, {
                $layoutType: "stack",
                $items: []
            }]
        }, {
            $layoutType: "columns",
            $layoutSubType: "50,50",
            $items: [{
                $layoutType: "stack",
                $items: []
            }, {
                $layoutType: "stack",
                $items: []
            }]
        }, {
            $layoutType: "columns",
            $layoutSubType: "50,50",
            $items: [{
                $layoutType: "stack",
                $items: []
            }, {
                $layoutType: "stack",
                $items: []
            }]
        }]
    },
    "composite-1-3": {
        $layoutType: "columns",
        $layoutSubType: "50,50",
        $items: [{
            $layoutType: "stack",
            $items: []
        }, {
            $layoutType: "stack",
            $items: [{
                $layoutType: "columns",
                $layoutSubType: "50,50",
                $items: [{
                    $layoutType: "stack",
                    $items: []
                }, {
                    $layoutType: "stack",
                    $items: []
                }]
            }, {
                $layoutType: "columns",
                $layoutSubType: "50,50",
                $items: [{
                    $layoutType: "stack",
                    $items: []
                }, {
                    $layoutType: "stack",
                    $items: []
                }]
            }, {
                $layoutType: "columns",
                $layoutSubType: "50,50",
                $items: [{
                    $layoutType: "stack",
                    $items: []
                }, {
                    $layoutType: "stack",
                    $items: []
                }]
            }]
        }]
    }
};

function CompositeLayout(){
}

exports.CompositeLayout = helpers.defineClass(CompositeLayout, Layout, {
    convertLayout: function(oldLayout){
        var $newLayout = this.box.$item.$layout = helpers.object.clone(_$layouts[this.$layout.$layoutSubType || "composite-3-1"], true);
        var newLayout = this.box.layoutContent = document.itemFactory.createLayout(this.box, $newLayout);
        newLayout.$$container = this.$$container || oldLayout.$$container;
        newLayout.render();
        newLayout.toggleAuthoring(true);
        layoutHelper.findFirstLayout(newLayout).convertLayout(oldLayout);
        var oldLayout = this;
        setTimeout(function(){
            document.controller.disposeObject(oldLayout);
        }, 100);
    }
});
