// syracuse-ui/lib/field/datetime/dateChoice.js
"use strict";
//http://www.openweb.eu.org/articles/formulaire_accessible
var helpers = require('syracuse-core').helpers;
var _monthYearChoice = require('syracuse-ui/lib/field/datetime/monthYearChoice');
var _globals = require('streamline-runtime').globals;


function _onChangePeriode() {
	var item = this.parent;
	if (this.period == "month") {
		item._currentDate = item._currentDate.addMonths(this.isPrev ? -1 : 1);
	} else
	if (this.period == "year") {
		item._currentDate = item._currentDate.addYears(this.isPrev ? -1 : 1);
	} else {
		item._currentDate = item._currentDate.addDays(this.isPrev ? -7 : 7);
	}
	item._drawBody();
}

function _onQuickItemClick() {
	var item = this.parent;
	item._currentDate = item._currentDate[(this.isBegin ? "begOf" : "endOf") + this.interval]((this.interval == "Week") ? 1 : undefined);
	item._drawBody();
	item.quickList.style.display = "none";
}

function _onQuickClick() {
	var item = this.parent;
	if (!item.quickList) {
		item.quickList = document.createElement("nav");
		item.quickList.style.display = "none";
		item.quickList.className = "s-calendar-quick-list";
		var intervals = ["Year", "Quarter", "Month", "Week"];
		for (var ii = 0, jj = intervals.length; ii < jj; ii++) {
			var row = syra_dom.text("s-calendar-quick", syra_local["fdpIntervalOf" + intervals[ii]]);
			row.syraInterval = intervals[ii];
			row.syraOnClick = "onQuickItemClick";
			row.insertBefore(syra_button.add({
				parent: item,
				slot: row,
				link: document.createElement("td"),
				text: syra_local.fdpEnd,
				css: "s-calendar-quick-link",
				click: _onQuickItemClick,
				interval: intervals[ii]
			}).link, row.firstChild);
			row.insertBefore(syra_button.add({
				parent: item,
				slot: row,
				link: document.createElement("td"),
				text: syra_local.fdpBegin,
				css: "s-calendar-quick-link",
				click: _onQuickItemClick,
				interval: intervals[ii],
				isBegin: true
			}).link, row.firstChild);
			item.quickList.appendChild(row);
		}
		item.domItem.appendChild(item.quickList);
	}
	if (item.quickList.style.display == "") {
		item.quickList.style.display = "none";
	} else {
		item.quickList.style.display = "";
		$(item.quickList).position({
			my: "left bottom",
			at: "right bottom",
			of: $(this.link)
		});
	}
}

function _onMonthClick() {
	var self = this;
	var item = self.parent;
	if (!item._month) {
		item._month = new _monthYearChoice.MonthChoice();
		item._month.create(item.field, item.domItem, item._currentDate, function(newDate) {
			item._currentDate = newDate;
			setTimeout(function() {
				item._drawBody();
				self.click();
			}, 200);
			return true;
		});
		item._month.toggle(true);
	} else {
		item._month.toggle(false);
		item._month.dispose();
		delete item._month;
	}
}

function _onYearClick() {
	var self = this;
	var item = self.parent;
	if (!item._year) {
		item._year = new _monthYearChoice.YearChoice(item.field);
		item._year.create(item.field, item.domItem, item._currentDate, function(newDate) {
			item._currentDate = newDate;
			setTimeout(function() {
				item._drawBody();
				self.click();;
			}, 200);
			return true;
		});
		item._year.toggle(true);
	} else {
		item._year.toggle(false);
		item._year.dispose();
		delete item._year;
	}
}

function _onDayClick(event, btn) {
	this.parent.field.setButtonValue(syra_culture.date.fromInternalValue(this.value).toString(this.parent.field.localFormat), this.value);
}


function DateChoice() {

}

exports.DateChoice = helpers.defineClass(DateChoice, null, {
	applyShortCut: function(shortcuts, event) {
		if (shortcuts.esc) {
			if (shortcuts.t) {
				this._currentDate = syra_culture.date.today();
				this._drawBody();
				return true;
			}
		}
		if (shortcuts.down || shortcuts.up || shortcuts.left || shortcuts.right) {
			if (event.shiftKey) {
				this._currentDate = this._currentDate.addMonths((shortcuts.up || shortcuts.left) ? -1 : 1);
			} else {
				if (event.ctrlKey) {
					this._currentDate = this._currentDate.addYears((shortcuts.up || shortcuts.left) ? -1 : 1);
				} else {
					if (shortcuts.down || shortcuts.up) {
						this._currentDate = this._currentDate.addDays(shortcuts.up ? -7 : 7);
					}
					if (shortcuts.left || shortcuts.right) {
						this._currentDate = this._currentDate.addDays(shortcuts.left ? -1 : 1);
					}
				}
			}
			this._drawBody();
			return true;
		}
		if (shortcuts.enter) {
			this._currentDateBtn && this._currentDateBtn.click();
			return true;
		}
	},
	create: function(field, options) {
		this.domItem = document.createElement("div");
		this.domItem.className = "s-calendar";
		this.domItem.style.display = "none";
		field.domItem.appendChild(this.domItem);
		if (options && (options.internalValue || options.internalValue === "")) {
			this._selectedDate = isNaN(options.internalValue) || options.internalValue === "" ? syra_culture.date.today() : syra_culture.date.fromInternalValue(options.internalValue);
			this._currentDate = syra_culture.date.fromInternalValue(this._selectedDate._value);
		} else {
			try {
				var value = field.getValue(); //catch exception if format error
				if (value) {
					this._selectedDate = syra_culture.date.parse(field.getInputValue(), field.localFormat);
				}
			} catch (error) {
				syra_fields.invalidate(field, error.message);
			}
			this._selectedDate = this._selectedDate || syra_culture.date.today();
			this._currentDate = syra_culture.date.fromInternalValue(this._selectedDate._value);
		}
		this.field = field;
		this._table = document.createElement("table");
		this._table.setAttribute("cellspacing", 0);
		this._table.className = "s-calendar-content";
		this._table.appendChild(this._addHead());
		this._table.appendChild(this.body = document.createElement("tbody"));
		this._table.appendChild(this._appendFoot());
		this.domItem.appendChild(this._table);

		this._drawBody();
	},
	remove: function() {
		syra_dom.remove(this.domItem);
		this.dispose();
	},
	_addHead: function() {
		var head = document.createElement("thead");
		var row = document.createElement("tr");
		var cell = document.createElement("th");
		cell.setAttribute("colspan", 8);
		cell.className = "s-calendar-month-year";

		// prev year link
		syra_button.add({
			parent: this,
			slot: cell,
			text: syra_local.fdpPrevYear,
			css: "s-calendar-btn",
			iconOnly: true,
			fontIcon: "first_step",
			shortCutTip: syra_shortCuts.tip.movePrevYear,
			click: _onChangePeriode,
			period: "year",
			isPrev: true
		});

		syra_button.add({
			parent: this,
			slot: cell,
			text: syra_local.fdpPrevMonth,
			css: "s-calendar-btn",
			iconOnly: true,
			fontIcon: "prev_step",
			shortCutTip: syra_shortCuts.tip.movePrevMonth,
			click: _onChangePeriode,
			period: "month",
			isPrev: true
		});

		this._monthBtn = syra_button.add({
			parent: this,
			slot: cell,
			text: syra_local.fdpChoiceMonth,
			css: "s-calendar-month-link",
			click: _onMonthClick
		});

		this._yearBtn = syra_button.add({
			parent: this,
			slot: cell,
			text: syra_local.fdpChoiceYear,
			css: "s-calendar-year-link",
			click: _onYearClick
		});

		syra_button.add({
			parent: this,
			slot: cell,
			text: syra_local.fdpNextMonth,
			css: "s-calendar-btn",
			iconOnly: true,
			fontIcon: "next_step",
			shortCutTip: syra_shortCuts.tip.moveNextMonth,
			click: _onChangePeriode,
			period: "month"
		});

		syra_button.add({
			parent: this,
			slot: cell,
			text: syra_local.fdpNextYear,
			css: "s-calendar-btn",
			iconOnly: true,
			fontIcon: "last_step",
			shortCutTip: syra_shortCuts.tip.moveNextYear,
			click: _onChangePeriode,
			period: "year"
		});



		row.appendChild(cell);
		head.appendChild(row);

		row = document.createElement("tr");
		var cell = document.createElement("th");
		cell.className = "s-calendar-week-day";
		row.appendChild(cell);

		var firstDayOfWeek = _globals.context.localePreferences.firstDayOfWeek;

		for (var ii = 0; ii < 7; ii++) {
			var cell = document.createElement("th");
			var dayOfWeek = (firstDayOfWeek + ii) % 7;
			cell.className = "s-calendar-week-day";
			cell.title = syra_culture.date.dayName(dayOfWeek);
			cell.textContent = syra_culture.date.dayName(dayOfWeek, true);
			row.appendChild(cell);
		}
		head.appendChild(row);
		return head;
	},
	_appendFoot: function() {
		var row = document.createElement("tr");
		var cell = document.createElement("td");
		cell.setAttribute("colspan", 3);
		cell.className = "s-calendar-foot-week";

		syra_button.add({
			parent: this,
			slot: cell,
			text: syra_local.fdpPrevWeek,
			css: "s-calendar-btn",
			iconOnly: true,
			fontIcon: "prev_step",
			shortCutTip: syra_shortCuts.tip.movePrevWeek,
			click: _onChangePeriode,
			period: "week",
			isPrev: true
		});


		var label = document.createElement("label");
		label.className = "s-calendar-foot-week-title";
		label.textContent = syra_local.fdpWeek;
		cell.appendChild(label);

		this._weekNumber = document.createElement("label");
		this._weekNumber.className = "s-calendar-foot-week-title-num";
		cell.appendChild(this._weekNumber);

		syra_button.add({
			parent: this,
			slot: cell,
			text: syra_local.fdpNextWeek,
			css: "s-calendar-btn",
			iconOnly: true,
			fontIcon: "next_step",
			shortCutTip: syra_shortCuts.tip.moveNextWeek,

			click: _onChangePeriode,
			period: "week"
		});

		row.appendChild(cell);

		cell = document.createElement("td");
		cell.setAttribute("colspan", 3);
		cell.className = "s-calendar-foot-today";

		syra_button.add({
			parent: this,
			slot: cell,
			text: syra_local.fdpToday,
			css: "s-calendar-today-link",
			shortCutTip: syra_shortCuts.tip.moveToday,
			click: function() {
				var selector = this.parent;
				selector._currentDate = syra_culture.date.today();
				selector.field.setButtonValue(syra_culture.date.fromInternalValue(selector._currentDate._value).toString(selector.field.localFormat), selector._currentDate._value);
			}
		});

		row.appendChild(cell);

		cell = document.createElement("td");
		cell.setAttribute("colspan", 2);
		cell.className = "s-calendar-foot-more";

		syra_button.add({
			parent: this,
			slot: cell,
			text: syra_local.fdpQuickSelection,
			css: "s-calendar-btn",
			iconOnly: true,
			fontIcon: "next_step",
			click: _onQuickClick
		});
		row.appendChild(cell);

		var foot = document.createElement("tfoot");
		foot.appendChild(row);
		return foot;
	},
	_drawBody: function() {
		var curMonth = this._currentDate.month;
		var month = syra_culture.date.monthName(curMonth);

		syra_button.setText(this._monthBtn, month);
		syra_button.setText(this._yearBtn, this._currentDate.year);

		syra_dom.empty(this.body);

		var curDate = syra_culture.date.fromInternalValue(this._currentDate._value);
		var begOfMonth = curDate = curDate.begOfMonth();
		var firstDayOfWeek = _globals.context.localePreferences.firstDayOfWeek;
		var firstWeekOfYear = _globals.context.localePreferences.firstWeekOfYear;

		curDate = curDate.begOfWeek(firstDayOfWeek);
		for (var weekRow = 0; weekRow < 6; weekRow++) {
			var row = document.createElement("tr");
			var weekDay = (weekRow == 0) ? begOfMonth : curDate;
			var cell = document.createElement("td");
			cell.className = "s-calendar-week-num";
			row.appendChild(cell);

			cell.textContent = weekDay.weekNumber(firstDayOfWeek, firstWeekOfYear);

			for (var day = 0; day < 7; day++) {
				var btn = syra_button.add({
					parent: this,
					slot: row,
					link: document.createElement("td"),
					text: curDate.day,
					css: "s-calendar-day-link",
					click: _onDayClick,
					value: curDate._value
				});
				if (curMonth != curDate.month) {
					btn.link.className += " s-calendar-other-month";
				}
				if (this._currentDate.equals(curDate)) {
					btn.link.className += " s-calendar-select";
					this._currentDateBtn = btn;
				}
				curDate = curDate.addDays(1);
			}
			this.body.appendChild(row);
		}
		this._weekNumber.textContent = this._currentDate.weekNumber(firstDayOfWeek, firstWeekOfYear);
	},
	dispose: function() {
		syra_site.disposeObject(this);
	}
}); // Dummy comment to clean up rollout repository #6948
