// syracuse-ui/lib/field/datetime/monthYearChoice.js
"use strict";
var helpers = require('syracuse-core').helpers;
var _dateApi = require('syracuse-core').types.date;

function _onClick() {
	var item = this.parent;
	if (item.selectedBtn) {
		item.selectedBtn.link.className = "s-calendar-my-item";
	}
	(item.selectedBtn = this).className = "s-calendar-my-item s-calendar-select";
	item._selectedValue = this.value;
	var month = (this.isMonth && item._selectedValue) ? item._selectedValue : item._dateValue.month;
	var year = (!this.isMonth && item._selectedValue) ? item._selectedValue : item._dateValue.year;
	item._close(item._dateValue = _dateApi.fromJsDate(new Date(year, month - 1, item._dateValue.day)));
}

function MonthChoice() {

}

exports.MonthChoice = helpers.defineClass(MonthChoice, null, {
	create: function(field, parent, dateValue, close) {
		this.id = field.id + "montchoice";
		syra_store.add(this);
		this.parent = parent;
		this.domItem = document.createElement("nav");
		this.domItem.className = "s-calendar-quick-list";
		this._close = close;
		this._dateValue = dateValue;
		var tbody = document.createElement("tbody");
		for (var ii = 0; ii < 6; ii++) {
			var row = document.createElement("tr");
			this.addBtn(row, ii + 1, _dateApi.monthName(ii + 1, true), this._dateValue.month);
			this.addBtn(row, ii + 7, _dateApi.monthName(ii + 7, true), this._dateValue.month);
			tbody.appendChild(row);
		}
		this.domItem.appendChild(tbody);
		this.domItem.style.display = "none";
		this.parent.appendChild(this.domItem);
	},
	addBtn: function(slot, value, text, selValue) {
		var btn = syra_menus.button.add({
			parent: this,
			slot: slot,
			link: document.createElement("td"),
			text: text,
			css: "s-calendar-my-item",
			btnclick: _onClick,
			isMonth: true,
			value: value
		});
		if (value == selValue) {
			btn.link.className += " s-calendar-select";
			this.selectedBtn = btn;
		}
	},
	toggle: function(show) {
		syra_dom.hide(this.domItem, !show);
		if (show) {
			$(this.domItem).position({
				my: "right top",
				at: "left top",
				of: $(this.parent)
			});
		}
	},
	dispose: function() {
		syra_store.remove(this);
		syra_site.disposeObject(this);
	}
});


function YearChoice() {

}

exports.YearChoice = helpers.defineClass(YearChoice, null, {
	create: function(field, parent, dateValue, close) {
		this.id = field.id + "yearchoice";
		syra_store.add(this);
		this.parent = parent;
		this.domItem = document.createElement("nav");
		this.domItem.className = "s-calendar-quick-list";
		this._close = close;
		this._dateValue = dateValue;
		this._selectedValue = this._dateValue.year;
		var tbody = document.createElement("tbody");
		this.yearBtns = [];

		for (var ii = 0; ii < 6; ii++) {
			var row = document.createElement("tr");
			if (ii == 0) {
				this.addPrevNextCell(row, true);
				this.addPrevNextCell(row, false);
			} else {
				var cuYear = this._dateValue.year + ii - 4;
				this.yearBtns.push(this.addBtn(row, cuYear));
				this.yearBtns.push(this.addBtn(row, cuYear + 5));
			}
			tbody.appendChild(row);
		}
		this.domItem.appendChild(tbody);
		this.domItem.style.display = "none";
		this.parent.insertBefore(this.domItem, this.parent.firstChild);
	},
	addPrevNextCell: function(row, isPrev) {
		var td = document.createElement("td");
		td.className = "s-calendar-my-link-cell";
		syra_menus.button.add({
			parent: this,
			slot: td,
			text: isPrev ? syra_local.fdpPrevYear : syra_local.fdpNextYear,
			css: "s-calendar-btn",
			iconOnly: true,
			fontIcon: isPrev ? "prev_step" : "next_step",
			btnclick: function(event, btn) {
				var step = this.isPrev ? (-10) : 10;
				var yearBtns = this.parent.yearBtns;
				for (var ii = 0, jj = yearBtns.length; ii < jj; ii++) {
					var btn = yearBtns[ii];
					btn.value += step;
					syra_menus.button.setText(btn, btn.value);
					if (btn.value == this.parent._selectedValue) {
						btn.link.className += " s-calendar-select";
						this.parent.selectedBtn = btn;
					}
				}
			},
			isPrev: isPrev
		});
		row.appendChild(td);
	},
	addBtn: function(slot, value) {
		var btn = syra_menus.button.add({
			parent: this,
			slot: slot,
			link: document.createElement("td"),
			text: value,
			css: "s-calendar-my-item",
			btnclick: _onClick,
			value: value
		});
		if (value == this._selectedValue) {
			btn.link.className += " s-calendar-select";
			this.selectedBtn = btn;
		}
		return btn;
	},
	toggle: function(show) {
		syra_dom.hide(this.domItem, !show);
		if (show) {
			$(this.domItem).position({
				my: "left top",
				at: "right top",
				of: $(this.parent)
			});
		}
	},
	dispose: function() {
		syra_store.remove(this);
		syra_site.disposeObject(this);
	}

});
