// syracuse-trace/lib/representations/trace.js
"use strict";

exports.representation = {
	$entityName: "trace",
	$facets: {
		$query: {
			$layout: {
				$items: [{
					$category: "section",
					$layout: {
						$items: [{
							$category: "section",
							$layout: {
								$items: [{
									$bind: "$resources",
									$layout: {
										$items: [{
											$bind: "stamp",
										}, {
											$bind: "tag"
										}, {
											$bind: "category",
										}, {
											$bind: "severity",
										}, {
											$bind: "message",
										}, {
											$bind: "file",
										}, {
											$bind: "stack",
										}, {
											$bind: "session",
										}, {
											$bind: "hostname",
										}, {
											$bind: "pid",
										}, {
											$bind: "viewer",
										}]
									}
								}]
							}
						}]
					}
				}]
			}
		},
		$details: {
			$layout: {
				$items: [{
					"$category": "section",
					$layout: {
						$items: [{
							$bind: "stamp",
						}, {
							$bind: "tag"
						}, {
							$bind: "category",
						}, {
							$bind: "severity",
						}, {
							$bind: "message",
						}, {
							$bind: "file",
						}, {
							$bind: "stack",
						}, {
							$bind: "session",
						}, {
							$bind: "hostname",
						}, {
							$bind: "pid",
						}]
					}
				}]
			}
		},
		$edit: {
			$copy: "$details"
		}
	}
};
