// syracuse-trace/lib/representations/traceRecord.js
"use strict";

exports.representation = {
	resources: function() {
		var locale = require('streamline-locale');
		var globals = require('streamline-runtime').globals;
		return locale.resources(module, globals.context && globals.context.sessionLocale)();
	},
	$entityName: "traceRecord",
	$facets: {
		$query: {
			$layout: {
				$items: [{
					$category: "section",
					$layout: {
						$items: [{
							$category: "section",
							$layout: {
								$items: [{
									$bind: "$resources",
									$layout: {
										$items: [{
											$bind: "name"
										}, {
											$bind: "stamp"
										}, {
											$bind: "change"
										}, {
											$bind: "hostname"
										}, {
											$bind: "pid"
										}, {
											$bind: "status"
										}, {
											$bind: "onlySid"
										}, {
											$bind: "isSessionConcerned"
										}, {
											$bind: "sessionId"
										}, {
											$bind: "ticket"
										}]
									}
								}]
							}
						}]
					}
				}]
			}
		}
	}
};
