// syracuse-collaboration/lib/representations/storage/document.js
exports.representation = {
	$entityName: "document",
	$facets: {
		$details: {
			$layout: {
                $items: [{
                    $layoutType: "columns",
                    $layoutSubType: "50,50",
                    $items: [{
                    	$items: [{
	                        $title: "{@DocumentMainSectionTitle}",
	                        $category: "section",
	                        $layout: {
								$items: [{
									$bind: "description"
								}, {
									$bind: "isReadOnly"
								}, {
									$bind: "content"
								}, {
									$bind: "volume"
								}, {
									$bind: "owner"
								}]
	                        }
                    	}, {
	                        $title: "{@DocumentInfoSectionTitle}",
	                        $category: "section",
	                        $layout: {
								$items: [{
									$bind: "documentType"
								}, {
									$bind: "documentDate"
								}, {
									$bind: "fileName"
								}]
	                        }
                    	}]
					}, {
                    	$items: [{
	                        $title: "{@DocumentOrganizationSectionTitle}",
	                        $category: "section",
	                        $layout: {
								$items: [{
									$bind: "tags"
								}, {
									$bind: "teams"
								}]
	                        }
                    	} /* , {
	                        $title: "{@DocumentInternalTagsSectionTitle}",
	                        $category: "section",
	                        $layout: {
								$items: [{
									$bind: "internalTags"
								}]
	                        }
                    	}*/]
					}]
				}]
			}
		},
		$edit: {
			$layout: {
				$copy: "$details"
			}
		},
        $query: {
            $layout: {
                $items: [{
                    $category: "section",
                    $layout: {
                        $items: [{
                            $bind: "$resources",
                            $layout: {
                                $items: [{
                                    $bind: "description"
                                }, {
                                    $bind: "documentType"
                                }, {
                                    $bind: "documentDate"
                                }, {
                                    $bind: "fileName"
                                }, {
                                    $bind: "isReadOnly"
                                }, {
                                    $bind: "content"
                                }, {
                                    $bind: "volume"
                                }, {
                                    $bind: "owner"
                                }]
                            }
                        }]
                    }
                }]
            }
        }
	}
}
