// syracuse-ui/lib/scroll/carousel.js
"use strict";

function Carousel(config) {
	this.page = config.page || syra_site;
	this.container = config.container;
	this.scrollview = config.scrollview;
	this.y_axis = config.axis != "x";
	this.container.className += " s-carousel-slot";
	this.scrollview.className += " s-carousel-view";
	syra_item.initialize(this.page, this);
	var axis = this.y_axis ? "top" : "left";
	this.up = syra_button.add({
		parent: this,
		slot: this.container,
		css: config.css + "-btn s-carousel-btn s-carousel-" + axis,
		axis: axis,
		click: this.click,
		mouseUpDown: this.updown
	});
	var axis = this.y_axis ? "bottom" : "right";
	this.down = syra_button.add({
		parent: this,
		slot: this.container,
		css: config.css + "-btn s-carousel-btn s-carousel-" + axis,
		axis: axis,
		click: this.click,
		mouseUpDown: this.updown
	});
	this.wheel(true);
	this.bindMouse(true);
}

exports.Carousel = Carousel;

Carousel.prototype.bindMouse = function(bind) {
	var self = this;
	if (self.container) {
		if (bind) {
			self.container.addEventListener("mousedown", self.on_mousedown = function(event) {
				self.updown(event);
			}, false);
			self.container.addEventListener("mouseup", self.on_mouseup = function(event) {
				self.updown(event);
			}, false);
		} else {
			self.on_mousedown && self.container.removeEventListener("mousedown", self.on_mousedown);
			self.on_mouseup && self.container.removeEventListener("mouseup", self.on_mouseup);
		}
	}
};
Carousel.prototype.resize = function() {
	this.setIncrement();
	this.checkButtons();
};

Carousel.prototype.updown = function(event) {
	var btn = (event.target == this.up.link) ? this.up : (event.target == this.down.link) ? this.down : null;
	if (btn && !btn.isDisabled && !btn.isHidden) {
		if (this.mouseTimer) {
			clearInterval(this.mouseTimer);
			delete this.mouseTimer;
		}
		if (event.type == "mousedown") {
			if (this._isClickDisabled) {
				delete this._isClickDisabled;
			}
			btn.click();
			this._clickCount = 1;

			this.mouseTimer = setInterval(function() {
				if (!this.disposed) {
					this._clickCount++;
					btn.click();
				}
			}, 500);
		} else {
			if (this._clickCount) {
				this._isClickDisabled = true;
			}
		}
	}
};
Carousel.prototype.wheel = function(on) {
	var self = this;
	if (self.scrollview) {
		if (on) {
			if (!self.handler_wheel) {
				self.handler_wheel = self.scrollview.addEventListener("wheel", self.handler_wheel = function(event) {
					if (self.y_axis) {
						if (self.increment) {
							if (event.deltaY > 0) {
								self.scrollview.scrollTop += self.increment;
							} else
							if (event.deltaY < 0) {
								self.scrollview.scrollTop -= self.increment;
							}
						} else {
							self.scrollview.scrollTop += event.deltaY;
						}
					} else {
						self.scrollview.scrollLeft += event.deltaX;
					}
					self.checkButtons();
				});
			}
		} else {
			self.handler_wheel && self.scrollview.removeEventListener("scroll", self.handler_wheel);
			delete self.handler_wheel;
		}
	}
};

Carousel.prototype.checkButtons = function() {
	var diff, value;
	if (this.y_axis) {
		diff = this.scrollview.scrollHeight - this.scrollview.clientHeight;
		value = Math.ceil(this.scrollview.scrollTop);
	} else {
		diff = this.scrollview.scrollWidth - this.scrollview.clientWidth;
		value = this.scrollview.scrollLeft;
	}
	syra_button.visibility(this.up, diff > 0);
	syra_button.visibility(this.down, diff > 0);
	syra_button.disable(this.up, value == 0);
	syra_button.disable(this.down, (value - diff) >= 0);
};


Carousel.prototype.click = function() {
	var carousel = this.parent;
	if (!carousel._isClickDisabled) {
		switch (this.axis) {
			case "top":
				carousel.scrollview.scrollTop -= carousel.increment;
				break;
			case "left":
				carousel.scrollview.scrollLeft -= carousel.increment;
				break;
			case "bottom":
				carousel.scrollview.scrollTop += carousel.increment;
				break;
			case "right":
				carousel.scrollview.scrollLeft += carousel.increment;
				break;
		}
		carousel.checkButtons();
	}
	delete carousel._isClickDisabled;
};

Carousel.prototype.setIncrement = function() {
	var rect = this.scrollview.childNodes[0].getBoundingClientRect();
	var style = window.getComputedStyle(this.scrollview.childNodes[0], null);
	this.increment = this.y_axis ? rect.height + parseFloat(style.getPropertyValue("margin-bottom")) : rect.width;
};

Carousel.prototype.dispose = function() {
	this.wheel(false);
	this.bindMouse(false);
	syra_site.disposeObject(this);
};
