// syracuse-ui/lib/scroll/carousel.js
"use strict";

function Carousel(context) {
	this.page = context.page || syra_site;
	this.view = context.view;
	this.content = context.content;
	this.y_axis = context.axis != "x";
	syra_item.initialize(this.page, this);
	var axis = this.y_axis ? "top" : "left";
	this.up = syra_button.add({
		parent: this,
		slot: this.view,
		css: context.css + " s-carousel s-carousel-" + axis,
		axis: axis,
		click: this.click,
		mouseUpDown: this.updown
	});
	var axis = this.y_axis ? "bottom" : "right";
	this.down = syra_button.add({
		parent: this,
		slot: this.view,
		css: context.css + " s-carousel s-carousel-" + axis,
		axis: axis,
		click: this.click,
		mouseUpDown: this.updown
	});
	this.wheel(true);
	this.bindMouse(true);
}

exports.Carousel = Carousel;

Carousel.prototype.bindMouse = function(bind) {
	var self = this;
	if (self.view) {
		if (bind) {
			self.view.addEventListener("mousedown", self.on_mousedown = function(event) {
				self.updown(event);
			}, false);
			self.view.addEventListener("mouseup", self.on_mouseup = function(event) {
				self.updown(event);
			}, false);
		} else {
			self.on_mousedown && self.view.removeEventListener("mousedown", self.on_mousedown);
			self.on_mouseup && self.view.removeEventListener("mouseup", self.on_mouseup);
		}
	}
};
Carousel.prototype.updown = function(event) {
	var btn = (event.target == this.up.link) ? this.up : (event.target == this.down.link) ? this.down : null;
	if (btn && !btn.isDisabled && !btn.isHidden) {
		if (this.mouseTimer) {
			clearInterval(this.mouseTimer);
			delete this.mouseTimer;
		}
		if (event.type == "mousedown") {
			if (this._isClickDisabled) {
				delete this._isClickDisabled;
			}
			btn.click();
			this._clickCount = 1;

			this.mouseTimer = setInterval(function() {
				if (!this.disposed) {
					this._clickCount++;
					btn.click();
				}
			}, 500);
		} else {
			if (this._clickCount) {
				this._isClickDisabled = true;
			}
		}
	}
};
Carousel.prototype.wheel = function(on) {
	var self = this;
	if (self.content) {
		if (on) {
			if (!self.handler_wheel) {
				self.handler_wheel = self.content.addEventListener("wheel", self.handler_wheel = function(event) {
					if (self.y_axis) {
						if (self.increment) {
							if (event.deltaY > 0) {
								self.content.scrollTop += self.increment;
							} else
							if (event.deltaY < 0) {
								self.content.scrollTop -= self.increment;
							}
						} else {
							self.content.scrollTop += event.deltaY;
						}
					} else {
						self.content.scrollLeft += event.deltaX;
					}
					self.checkButtons();
				});
			}
		} else {
			self.handler_wheel && self.content.removeEventListener("scroll", self.handler_wheel);
			delete self.handler_wheel;
		}
	}
};

Carousel.prototype.checkButtons = function() {
	var diff, value;
	if (this.y_axis) {
		diff = this.content.scrollHeight - this.content.clientHeight;
		value = Math.ceil(this.content.scrollTop);
	} else {
		diff = this.content.scrollWidth - this.content.clientWidth;
		value = this.content.scrollLeft;
	}
	syra_button.visibility(this.up, diff > 0);
	syra_button.visibility(this.down, diff > 0);
	syra_button.disable(this.up, value == 0);
	syra_button.disable(this.down, (value - diff) >= 0);
};


Carousel.prototype.click = function() {
	var carousel = this.parent;
	if (!carousel._isClickDisabled) {
		switch (this.axis) {
			case "top":
				carousel.content.scrollTop -= carousel.increment;
				break;
			case "left":
				carousel.content.scrollLeft -= carousel.increment;
				break;
			case "bottom":
				carousel.content.scrollTop += carousel.increment;
				break;
			case "right":
				carousel.content.scrollLeft += carousel.increment;
				break;
		}
		carousel.checkButtons();
	}
	delete carousel._isClickDisabled;
};

Carousel.prototype.validate = function() {
	this.setIncrement();
	this.checkButtons();
};
Carousel.prototype.setIncrement = function() {
	var rect = this.content.childNodes[0].getBoundingClientRect();
	var style = window.getComputedStyle(this.content.childNodes[0], null);
	this.increment = this.y_axis ? rect.height + parseFloat(style.getPropertyValue("margin-bottom")) : rect.width;
};

Carousel.prototype.dispose = function() {
	this.wheel(false);
	this.bindMouse(false);
	syra_site.disposeObject(this);
};

exports.add = function(context) {
	return new Carousel(context);
};
