// syracuse-ui/lib/scroll/list.js
"use strict";
var Scrollbar = require('syracuse-ui/lib/scroll/scrollbar').Scrollbar;

function List(config) {
	this.config = config || {};
	this.ul = syra_dom.ul("s-popup-primary-ul", this.ulSlot);
	if (this.config.scroll) {
		this.ulSlot = syra_dom.div("s-popup-primary-ul-slot");
		this.scrollBar = new Scrollbar({
			container: this.ulSlot,
			scrollview: this.ul
		});
	}
	this.ulSlot && this.ulSlot.appendChild(this.ul);
	this.config.parentNode && config.parentNode.appendChild(this.ulSlot || this.ul);
}

List.prototype.addLine = function() {
	return this._lastLine = syra_dom.li("s-popup-primary-li", this.ul);
};

List.prototype.addButton = function(btn) {
	btn.parent = btn.parent || syra_site;
	btn.slot = this.addLine();
	btn.css = btn.css || "s-popup-primary-btn-default";
	return syra_button.add(btn);
};
List.prototype.addItem = function(page, $item) {
	return page.addItem(this.addLine(), $item);
};
List.prototype.setScrollheight = function(height) {
	this.ul.style.maxHeight = height + "px";
};
List.prototype.scrollTo = function(node) {
	syra_site.scrollToItem(node, this.ul, true);
};
List.prototype.addSeparator = function(line, add) {
	if (line) {
		syra_dom.toggleClass(line, "s-popup-primary-sep", add);
	} else {
		if (this._lastLine) {
			this._lastLine.className += " s-popup-primary-sep"
		}
	}

};
List.prototype.hide = function(hide) {
	syra_dom.hide(this.ulSlot, hide);
};
List.prototype.resize = function() {
	this.scrollBar.resize();
};

List.prototype.dispose = function() {
	this.scrollBar && this.scrollBar.dispose();
	syra_site.disposeObject(this);
};

exports.List = List;
