// syracuse-ui/lib/scroll/more.js
"use strict";
var Scrollbar = require('syracuse-ui/lib/scroll/scrollbar').Scrollbar;

function More(list) {
	var self = this;
	self.list = list;
	self.slot = syra_dom.div(this.css + "-more-slot");
	self.ul = syra_dom.ul(self.list.css + "-more-ul", self.slot);
	self.scrollBar = new Scrollbar({
		container: self.slot,
		scrollview: self.ul
	});
	self._btn = self.list.addButton({
		parent: syra_site,
		text: syra_local.megaMenuMore,
		click: function() {
			if (!self._popup) {
				self._popup = syra_over.openPopup(syra_site, {
					content: {},
					slot: self.slot,
					autoCloseBoundary: self.slot,
					picker: self._btn.link,
					isParentPopup: true,
					position: {
						my: "left top",
						at: "left bottom",
						of: self._btn.link
					},
					onresize: function(maxHeight) {
						self.ul.style.maxHeight = maxHeight + "px"
						self.scrollBar.resize();
					},
					close: function() {
						self._popup = null;
					}
				});
			} else {
				self._popup.close();
			}
		}
	});
	self.empty = function() {
		syra_dom.empty(this.ul);
	};

	self.resize = function() {
		var maxW = this.list.ul.clientWidth + 1; //1 for little space
		if (this.lastWidth != maxW) {
			this.lastWidth = maxW;
			var itemCss = this.list.css + "-more-li";
			if (this.list.ul.scrollWidth > maxW) {
				if (!this._btn.slot.parentNode) {
					this.list.ul.appendChild(this._btn.slot);
				}
				while (this.list.ul.scrollWidth > maxW) {
					var item = this._btn.slot.previousSibling;
					if (!item) {
						break;
					}
					item.className += " " + itemCss;
					this.ul.insertBefore(item, this.ul.firstChild);
				}
			} else {
				while (this.ul.childNodes.length) {
					var node = this.ul.childNodes[0];
					node.className = node.className.replace(itemCss, "");
					this.list.ul.insertBefore(node, this._btn.slot);
					if (this.list.ul.scrollWidth > maxW) {
						node.className += " " + itemCss;
						this.ul.insertBefore(node, this.ul.firstChild);
						break;
					}
				}
			}
			this._btn.slot.style.visibility = this.ul.childNodes.length ? "visible" : "hidden";
		}

	};
	self.dispose = function() {
		syra_button.remove(this._btn);
		this.scrollBar && this.scrollBar.dispose();
		syra_site.disposeObjectt(this);
	};
}

exports.More = More;
