// syracuse-ui/lib/links/menus/picker.js
"use strict";
var _lookup = require("syracuse-ui/lib/links/menus/lookup");

exports.menus = {
	add: function(field) {
		return syra_button.addFieldPicker({
			parent: field,
			pickerId: "menus",
			fontIcon: "menus",
			text: syra_local.flActions,
			shortCutTip: (field.isField && !field.isArrayField) ? syra_shortCuts.tip.contextMenu : null,
			click: this.onClick
		});
	},
	onClick: function() {
		syra_menus.open({
			scope: this.parent,
			picker: this
		});
	},
	show: function(field) {
		if (field.picker_menus) {
			var show = field.$isEditMode || field.hasValue();
			syra_button.hide(field.picker_menus, !show);
			if (!field.picker_menus.link.parentNode && field.onAddMenuPicker) {
				field.onAddMenuPicker(field.picker_menus);
			}
		}
	}
};

exports.$localize = {
	add: function(field, $menu) {
		if (field.$item.$useLocalizePicker) {
			$menu.$isHidden = true;
		}
		if (field.$isEditMode && !field.picker_localize && field.$item.$useLocalizePicker) {
			syra_button.addFieldPicker({
				parent: field,
				pickerId: "$localize",
				fontIcon: "localize",
				text: field.$menus.$localize.$title,
				click: this.onClick
			});
		}
		$menu.$isPopup = true;
	},
	onClick: function() {
		syra_localizer.onMenuClick({
			scope: this.parent,
			menu: this.parent.$menus.$localize
		});
	}
};

exports.$tunnel = {
	add: function(field, $menu) {
		$menu.$title = syra_local.fieldTunnel;
		$menu.$isPopup = syra_site.isTabletDevice || (field.$isEditMode && field.$item.$inplace && !field.$item.$isFilterMode);
		if (!$menu.$isPopup && !field.picker_$tunnel && !field.$item.$inplace) {
			syra_button.addFieldPicker({
				parent: field,
				pickerId: "$tunnel",
				fontIcon: "tunnel",
				text: syra_local.fieldTunnel,
				shortCutTip: syra_shortCuts.tip.openTunnel,
				click: this.onClick
			});
			if (!field.$isDisabled && field.$isReadOnly) {
				syra_button.hide(field.picker_$tunnel, false);
			}
		}
	},
	onClick: function() {
		var field = this.parent;
		syra_router.executeMenu(field.$menus.$tunnel, field.isRecordArticle ? field : field.articleParent);
	}
};

exports.$lookup = {
	add: function(field, $menu) {
		if (field.$isEditMode && !field.picker_$lookup) {
			syra_button.addFieldPicker({
				parent: field,
				pickerId: "$lookup",
				fontIcon: field.$item.$useSummary ? ((field.$item.$popupRight) ? "right_arrow" : "expand_m") : "lookup",
				text: syra_local.fieldLookup,
				shortCutTip: syra_shortCuts.tip.openLookup,
				click: this.onClick
			});
		}
		$menu.$isPopup = false;
	},
	onClick: function() {
		_lookup.show(this.parent);
	}
};
