// syracuse-ui/lib/tests/messagebox/messageBoxTest.js
"use strict";
var helpers = require("syracuse-core/lib/helpers");
var TestPage = require("syracuse-ui/lib/tests/testPage").TestPage;
var inlineSettings = require("./messageBoxInlineSettings");
var popupSettings = require("./messageBoxPopupSettings");

function MessageBoxTestInline() {}

exports.MessageBoxTestInline = helpers.defineClass(MessageBoxTestInline, TestPage, {
	loadBox: function() {
		this.messageTypes = inlineSettings.messageTypes;
		this.$item.$layout = inlineSettings.$item.$layout;
		for (var ii = 0, jj = this.messageTypes.length; ii < jj; ii++) {
			this.$prototype.$links[this.messageTypes[ii].$type] = {
				$title: "Open popup"
			};
		}
		TestPage.prototype.loadBox.call(this);
		for (var ii = 0, jj = this.messageTypes.length; ii < jj; ii++) {
			var msg = this.messageTypes[ii];
			var section = this.idMap["messageBoxSection-" + msg.$type];
			if (msg.$type == "info") {
				section.domTitle.textContent += " ($isAutoClose = 1000)";
			}
			if (msg.$type == "fatal") {
				section.domTitle.textContent += " ($isAutoClose = true 4s)";
			}

			msg.layoutSlot = section.body;
			document.site.showMessage(msg);
			delete msg.layoutSlot;
		}
	},
	onMenuClick: function(menu) {
		for (var ii = 0, jj = this.messageTypes.length; ii < jj; ii++) {
			var msg = this.messageTypes[ii];
			if (msg.$type == menu.$bind) {
				if (msg.cbk && !msg.callback) {
					msg.callback = function(response) {
						if (response) {
							alert(response[msg.$type == "prompt" ? "prompt" : "$clientId"]);
						}
					};
				}
				document.site.showMessage(helpers.object.clone(msg, true));
				return false;
			}
		}
		return true;
	}
});

function MessageBoxTestPopup() {}

exports.MessageBoxTestPopup = helpers.defineClass(MessageBoxTestPopup, TestPage, {
	_initiliazeTestPage: function() {
		this.messageTypes = popupSettings.messageTypes;
		this.$item.$layout = popupSettings.$item.$layout;
		this.$prototype = popupSettings.$prototype;
	},
	onMenuClick: function(menu) {
		document.site.showMessage(this.messageTypes[parseInt(menu.$bind, 10) - 1]);
		// lose focus
		//menu.$$item.blur();
	}
});
