// syracuse-ui/lib/tests/messagebox/messageBoxInlineSettings.js
"use strict";
exports.$item = {
	"$layout": {
		"$items": [{
			"$layoutType": "row",
			"$items": [{
				"$layoutType": "stack",
				"$items": [{
					"$category": "section",
					"$clientId": "messageBoxSection-alert",
					"$title": "$type = 'alert'",
					"$layout": {
						"$items": [{
							"$category": "link",
							"$bind": "alert"
						}]
					}
				}, {
					"$category": "section",
					"$clientId": "messageBoxSection-fatal",
					"$title": "$type = 'fatal'",
					"$layout": {
						"$items": [{
							"$category": "link",
							"$bind": "fatal"
						}]
					}
				}, {
					"$category": "section",
					"$clientId": "messageBoxSection-warning",
					"$title": "$type = 'warning'",
					"$layout": {
						"$items": [{
							"$category": "link",
							"$bind": "warning"
						}]
					}
				}]
			}, {
				"$layoutType": "stack",
				"$items": [{
					"$category": "section",
					"$clientId": "messageBoxSection-info",
					"$title": "$type = 'info'",
					"$layout": {
						"$items": [{
							"$category": "link",
							"$bind": "info"
						}]
					}
				}, {
					"$category": "section",
					"$clientId": "messageBoxSection-question",
					"$title": "$type = 'question'",
					"$layout": {
						"$items": [{
							"$category": "link",
							"$bind": "question"
						}]
					}
				}, {
					"$category": "section",
					"$clientId": "messageBoxSection-prompt",
					"$title": "$type = 'prompt'",
					"$layout": {
						"$items": [{
							"$category": "link",
							"$bind": "prompt"
						}]
					}
				}]
			}]
		}]
	}
};

exports.messageTypes = [{
	$type: "alert",
	$title: "This is an alert",
	$message: "Hi, I am an alert message\n with carriage return"
}, {
	$type: "fatal",
	$isAutoClose: true,
	$title: "This is a fatal message",
	$message: "Hi, I am a fatal  message"
}, {
	$title: "This is a warning",
	$message: "Hi, I am a warning message",
	$type: "warning"
}, {
	$isAutoClose: 1000,
	$title: "This is an information",
	$message: "Hi, I am an information message",
	$type: "info"
}, {
	$title: "This is a question",
	$message: "Hi, I am a question",
	$type: "question",
	$buttons: "yesnocancel",
	cbk: true,
	callback: function(response) {
		alert(response.$clientId);
	}
}, {
	$type: "prompt",
	$title: "This is a prompt",
	$message: "Hi, I am a prompt",
	cbk: true,
	callback: function(response) {
		alert(response.prompt);
		return true;
	}
}];
