// syracuse-tablet/html/js/controls/page/ctrlVignetteFooter.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var Base = require('syracuse-tablet/html/js/controls/page/ctrlFooterBase').Klass;


/**
 * Vignette footer - USed to call action in dashboards
 * Created only for smartphone devices
 */
var _Klass = utils.defineClass(
	function CtrlVignetteFooter($type, vignette, options) {
		Base.call(this, $type, vignette.controller, options);
		this._vignette = vignette;
	}, Base, {

		destroy: function() {
			Base.prototype.destroy.call(this);
			this._vignette = null;
		},

		getFooterOptions: function(buildOptions) {
			return {
				showHome: false,
				showPageConfig: false,
				statusPinPage: false,
				multiSelectionOnly: this.options.multiSelectionOnly === true,
				// To force build
				forceRefresh: true
			};
		},

		getActionController: function() {
			if (!this._vignette) return null;
			return this._vignette.getPage();
		},

		/**
		 * Called after the footer has been attached
		 * Enable multiselection
		 */
		notifToggled: function(shown) {
			Base.prototype.notifToggled.call(this, shown);
			if (shown === true && this._multiSelManager) {
				this.toggleMultiSelection();
			}
		}

	});

exports.Klass = _Klass;
