// syracuse-tablet/html/js/controls/page/ctrlSidePanel.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var Base = require('syracuse-tablet/html/js/controls/ctrlBase').Klass;
var eventListener = require('syracuse-tablet/html/js/application/eventListener');
var notifications = require('syracuse-tablet/html/js/helpers/notifications');
var prototype = require('syracuse-tablet/html/js/helpers/prototype');

var _templates = {
	main: '\
			<nav class="navbar s-m-side-panel {{side}} navbar-inverse ctrl-evt-click" role="navigation">\
				<div class="container-fluid">\
					<div class="navbar-header">\
						<span class="navbar-brand" href="#"><span>{{title}}</span></a>\
					</div>\
				</div>\
				<ul class="nav navbar-nav">\
					{{{actionsHtml}}}\
				</ul>\
				<div class="s-m-side-panel-options">\
					<a draggable="false"  href="#" data-action="resize" data-control-id="{{ctrlId}}"><i class="fa fa-2x fa-bars"></i></a>\
				</div>\
			</nav>'
};
/**
 * Base left/right panel control for dashBoards and regular pages
 * 	Adapted from Metro Bootstrap navbar-side
 */
var _Klass = utils.defineClass(

	function(controller, $type, side, options) {
		if (["left", "right"].indexOf(side) < 0) throw new Error("side panel - Unexpected side");
		this.side = side;
		Base.call(this, controller, {
			$bind: side + "panel"
		}, prototype.create({
			$type: $type
		}), options);
	}, Base, {

		/**
		 * Just build the root elemt - see afterResizeBuildHtml
		 */
		buildHtml: function($$parent, controllerDao, buildOptions) {
			this.buildOptions = buildOptions.sidePanel || {};
			Base.prototype.buildHtml.call(this, $$parent, controllerDao, buildOptions, [this.typeName]);
			this.$$elmt.hide();
		},

		applyMetaData: function(metaData) {
			//  No meta for this control 
			return;
		},

		/**
		 * Build the html here when we know the orientation and deviceClass
		 */
		onMainPageResize: function(pageInfo, orientation, deviceClass) {
			var context = {
				ctrlId: this.id,
				side: this.side
			};
			this.afterResizeBuildHtml(context, pageInfo, orientation, deviceClass);
		},
		afterResizeBuildHtml: function(context, pageInfo, orientation, deviceClass) {
			this.$$panel = $(Handlebars.compile(_templates.main)(context));
			this.$$panel.appendTo(this.$$elmt);
		},

		/**
		 * Reduce/Enlarge the size of the panel
		 * Notify the page (sidePanel) with resize side,action
		 * 		sidePanelToggle is also called by footer with toggle action to show/hide the panel
		 * The page call the toggle action below to get the width and shift the panel
		 */
		_actResize: function() {
			var self = this;
			self.$$panel.toggleClass('closed');
			eventListener.triggerAction(self.$$elmt, "sidePanel", "resize," + self.side);
		},

		/**
		 * Notify the page that the panel is loaded to show/hide the action in footer
		 */
		publishReady: function() {
			var self = this;
			setTimeout(function() {
				notifications.publish("sm.side.panel.ready", self.$type);
			}, 10);
		},

		/**
		 * Toggle the panel
		 * 	action: toggle(click button footer), close(click dismiss) or resize (click resize icon inside the panel)
		 * 	callBack: shift the content for left panel
		 * 	This action is called by footer action and also by toggle data-action
		 */
		toggle: function(action, callBack) {
			var self = this;
			var e = self.$$elmt;
			// TODO - Do not call toggle if panel has not initialized - see test dashboards
			if (!e || !self.$$panel) return;
			var width = self.$$panel.width();
			// close -> Force hide - resize -> stay visible - toggle -> toggle
			e.toggle(action === "close" ? false : action === "resize" ? true : action === "toggle" ? undefined : undefined);
			if (callBack) callBack(e.is(':visible') === true ? width : 0);
		},
		onClick: function(evt) {
			this.close();
		},
		// Trigger evt to let page manage the close and all additional actions
		close: function() {
			eventListener.triggerAction(this.$$elmt, "sidePanel", "close," + this.side);
		}
	});

exports.Klass = _Klass;
