// syracuse-tablet/html/js/controls/page/ctrlPageHeader.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var Base = require('syracuse-tablet/html/js/controls/ctrlBase').Klass;
var globals = require('syracuse-tablet/html/js/helpers/globals');
var fontUtils = require('syracuse-tablet/html/js/ui/fontUtils');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var eventListener = require('syracuse-tablet/html/js/application/eventListener');
var prototype = require('syracuse-tablet/html/js/helpers/prototype');

var _templates = {
	main: '\
		<div class="s-m-header container-fluid {{typeDevice}}">\
			{{{contentHtml}}}\
		</div>\
	',
	tablet: {
		main: '\
			<div class="s-m-header-nav">\
				{{#if showBack}} \
					<a href="#" data-action="historyBack">\
						<span class="' + fontUtils.pageIcon("back") + '" />\
						</a>\
				{{/if}} \
				{{#if showHome}} \
					<a href="#" data-action="gotoWelcomeApplication">\
						<span class="' + fontUtils.pageIcon("home") + '" />\
					</a>\
				{{/if}} \
			</div>\
			<div class="s-m-header-labels">\
				<div class="s-m-header-brand"><span class="s-m-brand">{{labelBrand}}</span> <span class="s-m-product">{{labelProduct}}</span></div>\
				<div class="s-m-header-appname">{{appName}}</div>\
			</div>\
			<div class="s-m-header-links">\
				{{{userContent}}}\
			</div>',
		userInfo: '\
			<div>\
				<span class="s-m-nav-text username">{{username}}</span>\
				<span class="s-m-nav-text userrole">{{userrole}}</span>\
			</div>\
			{{#if userPhotoUrl}}\
				<img src="{{userPhotoUrl}}"/>\
			{{else}}\
				<span style="font-size:32px" class="' + fontUtils.pageIcon("switchContext") + '" />\
			{{/if}}',
		infoOnly: '\
			<div>{{{userInfo}}}</div>}}}',
		dropDown: '\
			<div class="dropdown">\
				<a href="#" data-toggle="dropdown" role="button" id="{{dropdownId}}" aria-haspopup="true" aria-expanded="false">\
					{{{userInfo}}}\
				</a>\
				{{{dropDownMenu}}}\
			</div>',
		dropDownMenu: '\
			<ul class="dropdown-menu" role="menu" aria-labelledby="{{dropdownId}}">\
				{{#each links}}\
					{{#if divider}} \
						<li role="presentation" class="divider"></li>\
					{{/if}} \
					<li>\
						<a href="#" data-action="linkClicked" data-params="{{data-params}}" data-control-id="{{ctrlId}}" class="list-group-item {{css}}">\
							<span class="{{icon}}" />\
							<span>{{label}}</span>\
						</a>\
					</li>\
				{{/each}}\
			</ul>',
		search: '\
			<div class="s-m-header-search" style="display:none">\
				<form style="display:none" class="navbar-form navbar-right" role="search">\
					<div class="form-group input-group-sm"">\
						<input type="text" class="form-control" placeholder="Search">\
					</div>\
					<button type="submit" class="btn btn-success">\
						<span class="' + fontUtils.pageIcon("search") + '"></span>\
					</button>\
				</form>\
			</div>'
	},
	smartphone: {
		main: '\
			<div class="s-m-header-nav">\
				{{{backDom}}}\
				<a href="#" data-action="sidePanel" data-params="toggle,left">\
					<span class="' + fontUtils.pageIcon("bars") + '" />\
				</a>\
			</div>\
			<div class="s-m-header-labels">\
				{{{labelContent}}}\
			</div>\
			<div class="s-m-header-links">\
				{{{linksContent}}}\
			</div>',
		backDom: '\
			<a href="#" data-action="historyBack">\
				<span class="' + fontUtils.pageIcon("backIphone") + '" />\
			</a>',
		brandDom: '\
			<div class="s-m-header-brand">\
				<span class="s-m-brand">{{labelBrand}}</span>\
				<span class="s-m-product">{{labelProduct}}</span>\
			</div>',
		labelDom: '\
			<span>{{label}}</span>'
	}
};

var _linksCtx = {
	showAuthoring: {
		items: [{
			"data-params": "designPage",
			"icon": "designPage",
			"label": "header.links.designPage"
		}]
	},
	showLinks: {
		items: [{
			"data-params": "clearCache",
			"icon": "clearCache",
			"label": "header.links.clearCache"
		}, {
			"data-params": "switchContext",
			"icon": "switchContext",
			"label": "header.links.switchContext"
		}, {
			"data-params": "about",
			"icon": "about",
			"label": "header.sm.links.about"
		}]
	},
	showDevelopment: {
		"divider": true,
		items: [{
			"data-params": "developmentPage",
			"icon": "developmentPage",
			"label": "header.links.development"
		}]
	}
};
var _getHtml = function(type, name, ctx) {
	type = type || "";
	var key = type + name + "compiled";
	var tmpl = _templates[key];
	if (!tmpl) {
		tmpl = type.length == 0 ? _templates[name] : _templates[type][name];
		tmpl = _templates[key] = Handlebars.compile(tmpl);
	}
	return tmpl(ctx);
};

var _$proto = {
	$type: "tablet/x-pageheader"
};
var _$article = {
	$bind: "header"
}; /** Header height */
var _heights = {
	"tablet": 50,
	"smartphone": 50
};
/**
 * Page header Generate Html
 */
var _Klass = utils.defineClass(

	function(controller, options) {
		Base.call(this, controller, _$article, prototype.create(_$proto), options);
	}, Base, {

		destroy: function() {
			Base.prototype.destroy.call(this);
		},
		buildHtml: function($$parent, controllerDao, buildOptions) {
			this.buildOptions = buildOptions.header || {};
			this.buildOptions.showDevelopment = globals.isDvlpMode();
			Base.prototype.buildHtml.call(this, $$parent, controllerDao, buildOptions, [this.typeName]);
			// Built is done in onMainPageResize
		},
		applyMetaData: function(metaData) {
			//  No meta for this control 
			return;
		},
		_actLinkClicked: function(actionName) { /* triggering action */
			eventListener.triggerAction(this.$$elmt, actionName);
		},
		onMainPageResize: function(info, orientation, deviceClass) {
			var $$parent = this.$$elmt.parent();
			this._height = _heights[globals.getSiteLayout().getDeviceType(deviceClass)];
			$$parent.css({
				left: 0 + "px",
				width: info.contentWidth + "px",
				height: this._height + "px"
			});
			this._build(info, orientation, deviceClass);
		},
		_build: function(info, orientation, deviceClass) {
			if (this.$$header) this.$$header.remove();
			var typeDevice = globals.getSiteLayout().getDeviceType(deviceClass);
			var opts = this.buildOptions;
			var contentHtml;
			if (typeDevice === "smartphone") {
				contentHtml = this._smartPhoneHtml(opts);
			} else {
				contentHtml = this._tabletHtml(opts);
			}
			this.$$header = $(_getHtml(null, "main", {
				contentHtml: contentHtml,
				typeDevice: typeDevice
			}));
			this.$$header.appendTo(this.$$elmt);
		},
		/**
		 * Returns HTML for tablet
		 */
		_tabletHtml: function(info, orientation, deviceClass) {
			var self = this;
			var lknCtx = {
				dropdownId: utils.UUID(),
				links: []
			};
			var opts = self.buildOptions;
			var linksCpt = 0;
			for (var p in _linksCtx) {
				linksCpt++;
				if (opts[p] === true) {
					var l = _linksCtx[p];
					l.items.forEach(function(x, idx) {
						x = $.extend({}, x);
						x.label = locale.text(x.label);
						x.ctrlId = self.id;
						x.icon = fontUtils.pageIcon(x.icon);
						if (idx === 0 && linksCpt > 1) x.divider = l.divider;
						if (x["data-params"] == "clearCache" && globals.getCache().isDisabled()) {
							x.css = "disabled";
						} else {
							x.css = "";
						}
						lknCtx.links.push(x);
					});
				}
			}
			var _mainCtx = {
				showBack: opts.showHome,
				showHome: opts.showHome,
				labelBrand: opts.labelBrand || "Sage",
				labelProduct: opts.labelProduct || "ERP X3",
				appName: globals.getApplication().getTitle()
			};
			var htmlUserInfo = _getHtml("tablet", "userInfo", {
				username: globals.getUserName(),
				userrole: globals.getUserRole(),
				userPhotoUrl: globals.getUserPhotoUrl
			});
			if (lknCtx.links.length === 0) {
				_mainCtx.userContent = _getHtml("tablet", "infoOnly", {
					userInfo: htmlUserInfo
				});
			} else {
				_mainCtx.userContent = _getHtml("tablet", "dropDown", {
					userInfo: htmlUserInfo,
					dropDownMenu: _getHtml("tablet", "dropDownMenu", lknCtx)
				});
				self.$$elmt.css({
					overflow: "visible"
				});
			}
			return _getHtml("tablet", "main", _mainCtx);
		},
		/**
		 * Returns HTML for smartphone
		 */
		_smartPhoneHtml: function(opts) {
			var self = this;
			var backDom = "";
			var labelContent = "";
			var linksContent = "";

			// build backDom only if iPhone
			if (opts.showHome && globals.getLookAndFeel() === "ios") {
				backDom = _getHtml("smartphone", "backDom", {});
			}

			// build label content. 
			labelContent = self._getLabelContent(opts);

			// build links content.
			linksContent = self._getLinksContent();

			var ctx = {
				backDom: backDom,
				labelContent: labelContent,
				linksContent: linksContent
			};

			return _getHtml("smartphone", "main", ctx);
		},
		/**
		 *	/!\ SMARTPHONE ONLY
		 *	Returns header label content which depends on the current page.
		 *	Label depends whether user is located in home page, app dashboard page, inside an app, or anywhere else
		 */
		_getLabelContent: function(opts) {
			var self = this;
			var labelContent;

			function getContent(isLabelDom, labelVal, opts) {
				if (isLabelDom) {
					return _getHtml("smartphone", "labelDom", {
						label: labelVal
					});
				}
				// brand name
				else {
					return _getHtml("smartphone", "brandDom", {
						labelBrand: opts.labelBrand || "Sage",
						labelProduct: opts.labelProduct || "ERP X3",
					});
				}
			}

			// home page --> brand name
			if (this.controller.state.isWelcomePage) {
				labelContent = getContent(false, null, opts);
			} else {
				// app dashboard page --> app title
				if (self.controller.state.isHomeDashboard) {
					labelContent = getContent(true, globals.getApplication().getTitle());
				} else {
					// inside app --> page title
					if (self.controller.prototype.data("$title")) {
						labelContent = getContent(true, self.controller.prototype.data("$title"));
					}
					// other cases (for example html static page)
					else {
						labelContent = getContent(false, null, opts);
					}
				}

			}
			return labelContent;
		},
		/**
		 *	/!\ SMARTPHONE ONLY
		 *	Returns header links content which depends on the current page.
		 *	- No link on welcome page and application dashboard
		 *	- On query page, links will be "create" and an anchor to open the actions panel (on the right)
		 *	- On detail page, links will be "edit" and an anchor to open the actions panel (on the right)
		 *	- On edit page, links will be "save", "cancel" and an anchor to open the actions panel (on the right)
		 *	Some actions will be displayed in the footer
		 */
		_getLinksContent: function() {

		},
		getHeight: function(deviceClass) {
			return this._height == null ? _heights["tablet"] : this._height;
		}
	});

exports.Klass = _Klass;
