// syracuse-tablet/html/js/controls/page/ctrlBreadCrumbs.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var Base = require('syracuse-tablet/html/js/controls/ctrlBase').Klass;
var globals = require('syracuse-tablet/html/js/helpers/globals');
var prototype = require('syracuse-tablet/html/js/helpers/prototype');
var uiSettings = require('syracuse-tablet/html/js/helpers/uiSettings');
var notifications = require('syracuse-tablet/html/js/helpers/notifications');

var _templates = {
	items: '\
		<div class="s-m-items">\
		{{#each items}}{{{html}}}{{/each}}\
		</ul>\
		',
	item: '<a draggable="false" href="#" data-action="breadcrumb-click" data-control-id="{{ctrlId}}" data-params="{{data}}"><div class="s-m-item"><div class="s-m-text">{{text}}</div>{{#if notLast}}<div class="fa fa-caret-right s-m-icon-small"></div>{{/if}}</div></a>'
};

var _getHtml = function(name, ctx) {
	var key = name + "compiled";
	var tmpl = _templates[key];
	if (!tmpl) {
		tmpl = _templates[name];
		tmpl = _templates[key] = Handlebars.compile(tmpl);
	}
	return tmpl(ctx);
};

var _Klass = utils.defineClass(
	function CtrlBreadCrumbs($type, controller, options) {
		Base.call(this, controller, {
			$bind: "breadcrumbs"
		}, prototype.create({
			$type: $type
		}), options);

		notifications.subscribe(this, ["sm.mainpage.loaded"]);
	}, Base, {

		destroy: function() {
			Base.prototype.destroy.call(this);
			this._pageStates = [];
		},

		buildHtml: function($$parent, controllerDao, buildOptions) {
			this._height = uiSettings.getProp("breadcrumbs.height");
			Base.prototype.buildHtml.call(this, $$parent, controllerDao, buildOptions, [this.typeName]);
			this.$$elmt.css("height", this._height);
		},
		onMainPageResize: function(info, orientation) {
			var deviceType = globals.getSiteLayout().getDeviceType();
			this.$$elmt.parent().toggle(deviceType === "tablet");
		},
		afterRender: function(updateLayout, buildOpts) {},
		applyMetaData: function(metaData, buildOptions) {
			//  No meta for this control 
			return;
		},
		getHeight: function() {
			return this.$$elmt.is(":visible") ? this._height : 0;
		},
		/**
		 * Needed after authoring update
		 */
		refresh: function() {
			this._buildBreadcrumbs();
		},
		notifMainpageLoaded: function() {
			this._buildBreadcrumbs();
		},
		_buildBreadcrumbs: function() {
			var history = globals.getApplication().history;
			var self = this;
			if (!self.$$elmt) {
				return;
			}
			var items = this._getItems(history);
			var crumbs = [];
			self._pageStates = [];
			$.each(items, function(idx, state) {
				var uuid = utils.UUID();
				self._pageStates[uuid] = $.extend(true, {}, state, {
					stackNum: idx
				});
				crumbs.push({
					html: _getHtml("item", {
						text: state.$title,
						notLast: (idx + 1 < items.length),
						ctrlId: self.id,
						data: uuid
					})
				});
			});
			var html = _getHtml("items", {
				items: crumbs
			});

			this.$$elmt.empty();
			this.$$elmt.append(html);
		},
		_getItems: function(history) {
			history = history || [];
			history = history.slice(0);
			history = history.reverse(); // Latest is first
			var stackToHome = [];

			// Filter history to only show stack until we reach welcome page for the first time to keep it as small as possible
			history.some(function(state) {
				stackToHome.push(state);
				if (state.isWelcomePage) {
					return true;
				}
				return false;
			});
			return stackToHome;
		},
		_actBreadcrumbClick: function(uuid) {
			var self = this;
			var state = self._pageStates[uuid];
			if (!state) return;
			var num = state.stackNum;
			if (num <= 0) return; // Click on current page, no need to do an action
			globals.getApplication().goBack(function PageFilter(histState) {
				if (num === 0) {
					return true;
				}
				num--;
			});
		}
	});

exports.Klass = _Klass;
