// syracuse-tablet/html/js/controls/page/ctrlHeaderBase.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var uiSettings = require('syracuse-tablet/html/js/helpers/uiSettings');
var Base = require('syracuse-tablet/html/js/controls/ctrlBase').Klass;
var globals = require('syracuse-tablet/html/js/helpers/globals');
var fontUtils = require('syracuse-tablet/html/js/ui/fontUtils');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var eventListener = require('syracuse-tablet/html/js/application/eventListener');
var prototype = require('syracuse-tablet/html/js/helpers/prototype');
var actionManager = require('syracuse-tablet/html/js/application/actionManager');
var jsutils = require('syracuse-tablet/html/js/helpers/jsutils');
var multiSelectionMgr = require('syracuse-tablet/html/js/application/multiSelectionMgr');
var notifications = require('syracuse-tablet/html/js/helpers/notifications');
var native = require('syracuse-tablet/html/js/helpers/native/native');

var _userPanelId = globals.PANELUSERSMARTPHONE;
var _actionPanelId = globals.PANELACTIONSREGULAR;

var _templates = {
	main: '\
		<div class="s-m-header container-fluid {{deviceType}}">\
			{{{contentHtml}}}\
		</div> \
	',
	tablet: {
		main: '\
			<div class="s-m-header-nav">\
				{{#if showBack}} \
					<a href="#" draggable="false" data-action="historyBack">\
						<i class="' + fontUtils.pageIcon("back") + '" />\
						</a>\
				{{/if}} \
			</div>\
			<div class="s-m-header-labels">\
				{{#if showBrand}}\
					<div class="s-m-header-brand"><span class="s-m-brand">{{labelBrand}}</span> <span class="s-m-product">{{labelProduct}}</span></div>\
				{{/if}}\
				<div class="s-m-header-label">{{headerLabel}}</div>\
			</div>\
			<div class="s-m-header-links">\
				{{{userContent}}}\
			</div>',
		userInfo: '\
			<div>\
				<span class="s-m-nav-text username">{{username}}</span>\
				<span class="s-m-nav-text userrole">{{userrole}}</span>\
			</div>\
			<span style="font-size:32px; vertical-align: 5px;" class="' + fontUtils.pageIcon("switchContext") + '" />\
			',
		infoOnly: '\
			<div>{{{userInfo}}}</div>',
		dropDown: '\
			<div class="dropdown">\
				<a href="#" draggable="false" data-toggle="dropdown" role="button" id="{{dropdownId}}" aria-haspopup="true" aria-expanded="false">\
					{{{userInfo}}}\
				</a>\
				{{{dropDownMenu}}}\
			</div>',
		dropDownMenu: '\
			<ul class="dropdown-menu" role="menu" aria-labelledby="{{dropdownId}}">\
				{{#each links}}\
					{{#if divider}} \
						<li role="presentation" class="divider"></li>\
					{{/if}} \
					<li>\
						<a href="#" data-action="linkClicked" draggable="false" data-params="{{data-params}}" data-control-id="{{ctrlId}}" class="list-group-item {{css}}">\
							<span class="{{icon}}" />\
							<span>{{label}}</span>\
						</a>\
					</li>\
				{{/each}}\
			</ul>',
		search: '\
			<div class="s-m-header-search" style="display:none">\
				<form style="display:none" class="navbar-form navbar-right" role="search">\
					<div class="form-group input-group-sm"">\
						<input type="text" class="form-control" placeholder="Search">\
					</div>\
					<button type="submit" class="btn btn-success">\
						<span class="' + fontUtils.pageIcon("search") + '"></span>\
					</button>\
				</form>\
			</div>'
	},
	smartphone: {
		main: '\
			<div class="s-m-header-nav">\
				{{#if showUserpanel}}\
					<a href="#" draggable="false" data-action="headerAction" data-params="{{panelId}}" data-control-id="{{ctrlId}}">\
						<span class="' + fontUtils.pageIcon("bars") + '" />\
					</a>\
				{{/if}} \
				{{{backDom}}}\
			</div>\
			<div class="s-m-header-labels">\
				{{{labelContent}}}\
			</div>\
			<div class="s-m-header-links">\
				{{{linksContent}}}\
			</div>',
		backDom: '\
			<a href="#" draggable="false" data-action="historyBack" style="padding-left:0;">\
				<span class="' + fontUtils.pageIcon("backIphone") + '" />\
			</a>',
		brandDom: '\
			<div class="s-m-header-brand">\
				<span class="s-m-brand">{{labelBrand}}</span>\
				<span class="s-m-product">{{labelProduct}}</span>\
			</div>',
		labelDom: '\
			<div class="s-m-header-label">{{label}}</div>',
		actionLink: '\
			<a href="#" draggable="false" class="s-m-link {{css}}" {{#if $isHidden}}style="display:none"{{/if}} data-action="{{#if action}}{{action}}{{else}}headerLink{{/if}}" data-params="{{$uuid}}" data-control-id="{{ctrlId}}">\
				<i class="{{icon}}"></i>\
			</a>',
		actionPanelToggle: '\
			<a href="#" class="s-m-link" data-action="headerAction" draggable="false" data-params="{{panelId}}" data-control-id="{{ctrlId}}">\
				<i class="' + fontUtils.pageIcon("sidePanelR") + '" />\
				<span>{{{title}}}</span>\
			</a>'
	}
};

var _linksCtx = {
	showAuthoring: {
		items: [{
			"data-params": "designPage",
			"icon": "designPage",
			"label": "header.links.designPage"
		}]
	},
	showPageShare: {
		items: [{
			"data-params": "pageShare",
			"icon": "nativeShare",
			"label": "header.links.page.share"
		}]
	},
	showUserMenuLinks: {
		items: [{
			"data-params": "switchContext",
			"icon": "switchContext",
			"label": "header.links.switchContext"
		}, {
			"data-params": "settings",
			"icon": "settings",
			"label": "header.links.settings"
		}, {
			"data-params": "about",
			"icon": "about",
			"label": "header.sm.links.about"
		}]
	},
	showDevelopment: {
		"divider": true,
		items: [{
			"data-params": "developmentPage",
			"icon": "developmentPage",
			"label": "header.links.development"
		}]
	}
};

var _getHtml = function(type, name, ctx) {
	type = type || "";
	var key = type + name + "compiled";
	var tmpl = _templates[key];
	if (!tmpl) {
		tmpl = type.length == 0 ? _templates[name] : _templates[type][name];
		tmpl = _templates[key] = Handlebars.compile(tmpl);
	}
	return tmpl(ctx);
};

var _hideSoftwareBackButton;
var _mustHideSoftwareBackButton = function() {
	if (_hideSoftwareBackButton != null) {
		return _hideSoftwareBackButton;
	}
	_hideSoftwareBackButton = native.hasCapability("hardwareBackButton");
};

var _hideDevFeatures;
var _mustHideDevFeatures = function() {
	if (_hideDevFeatures != null) {
		return _hideDevFeatures;
	}
	_hideDevFeatures = !native.isTestMode() && native.hasCapability("nativeApp");
};

/**
 * Page header Generate Html
 */
var _Klass = utils.defineClass(

	function CtrlPageHeader($type, controller, options) {
		Base.call(this, controller, {
			$bind: "header"
		}, prototype.create({
			$type: $type
		}), options);
	}, Base, {

		destroy: function() {
			this.resetSmartphonePanels();
			if (this._actionManager) {
				this._actionManager.destroy();
				this._actionManager = null;
			}
			if (this._multiSelManager) {
				this._multiSelManager.destroy();
				this._multiSelManager = null;
			}
			// end
			Base.prototype.destroy.call(this);
		},
		getPageControlType: function() {
			return "header";
		},
		resetSmartphonePanels: function() {
			if (this.controller) {
				// Detach the panels from main page - Used by vignette header for smartphone
				this.controller.panelDetach("right");
				this.controller.panelDetach("left");
			}
			if (this._smphUserpanel) {
				this._smphUserpanel.destroy();
				this._smphUserpanel = null;
			}
			if (this._smphPageactionpanel) {
				this._smphPageactionpanel.destroy();
				this._smphPageactionpanel = null;
			}
		},
		buildHtml: function($$parent, controllerDao, buildOptions) {
			buildOptions = buildOptions || {};
			Base.prototype.buildHtml.call(this, $$parent, controllerDao, buildOptions, [this.typeName]);
			this.headerOptions = this.getHeaderOptions(buildOptions);
			// Build the header with default deviceClass - To avoid flickering of the page - Improvement is done in onMainPageResize
			if (buildOptions.refresh === true || this.headerOptions.forceRefresh) {
				// Avoid double build
				// If it's not a refresh build is called on onResize - there's always a triggerResize when we rendre a page
				this._build();
			}
		},

		// Can be overridden
		getHeaderOptions: function(buildOptions) {
			var opts = $.extend(buildOptions.footer, buildOptions.toolBars) || {};
			opts.showDevelopment = (globals.isMobile() || _mustHideDevFeatures() === true) ? false : globals.isDvlpMode();
			opts.showAuthoring = (utils.isDesktop() === true && _mustHideDevFeatures() === false) ? opts.showAuthoring : false; // Show full feature authoring only if on PC
			// Authoring of chart detail page needs to be done in parent dashboard
			if (this.controller.state.type === "chartdetail") {
				opts.showAuthoring = false;
			}
			return opts;
		},

		applyMetaData: function(metaData, buildOptions) {
			//  No meta for this control 
			return;
		},
		onMainPageResize: function(info, orientation) {
			// In any case because we need to have the content width	
			this.$$elmt.parent().css({
				width: info.contentWidth + "px"
			});
			this._build();
			// In any case because we need to have the width tho calculate text-overflow elipsis
			this._updateHeaderStyle(this._deviceType);
		},
		getHeight: function() {
			return this.$$elmt.is(":visible") ? this._height : 0;
		},
		/**
		 * Allows to build header in build html (better UX)
		 */
		_build: function() {
			var deviceType = globals.getSiteLayout().getDeviceType();
			if (this._deviceType != null && this._deviceType === deviceType) return;
			if (this._multiSelManager) {
				this._multiSelManager.destroy();
				this._multiSelManager = null;
			}
			if (this._actionManager) {
				this._actionManager.destroy();
				this._actionManager = null;
			}
			this._deviceType = deviceType;
			if (this.$$header) this.$$header.remove();
			this._height = uiSettings.getProp("header.height");
			this.$$elmt.css({
				height: this._height + "px"
			});
			/*this.$$elmt.height(h);*/
			var $$parent = this.$$elmt.parent();
			$$parent.css({
				left: 0 + "px",
				height: this._height + "px"
			});
			var opts = this.headerOptions;
			var labels = locale.getProductLabels();
			opts.labelBrand = labels.brand;
			opts.labelProduct = labels.product;
			// Controller that handle action - different from current page for vignettes
			var actionController = this.getActionController();
			// Controller that displays the footer
			var displayController = this.controller;
			var contentHtml;
			if (deviceType === "smartphone") {
				contentHtml = this._smartPhoneHtml(opts, displayController, actionController);
			} else {
				contentHtml = this._tabletHtml(opts, displayController, actionController);
			}
			this.$$header = $(_getHtml(null, "main", {
				contentHtml: contentHtml,
				deviceType: deviceType
			}));
			this.$$header.appendTo(this.$$elmt);
		},

		/**
		 * Controller that handles actions
		 * For actions in dashboard this controller is the page nested in the vignette
		 * Can be overridden
		 */
		getActionController: function() {
			return this.controller;
		},

		/**
		 *	Update some header css properties.
		 */
		_updateHeaderStyle: function(deviceType) {
			// tablet
			// calculate and set s-m-header-labels div width according
			// to widths of s-m-header-nav and s-m-header-links div elements.
			var self = this;
			var $$header = self.$$elmt.find(".s-m-header");
			var $$links = self.$$elmt.find(".s-m-header-links");
			var $$nav = self.$$elmt.find(".s-m-header-nav");
			var $$label = self.$$elmt.find(".s-m-header-label");
			var offset = 10;
			if (deviceType === "tablet") {

				// for pages other than welcome page, update label width
				if (!self.controller.state.isWelcomePage) {
					var maxWidth = $$header.width() - ($$nav[0].scrollWidth + $$links[0].scrollWidth + offset);
					$$label.css("max-width", maxWidth);
				}
				// on welcome page, if header is too small, label will be hidden for a better display
				else {
					$$label.css("max-width", "inherit");
					var $$brand = self.$$elmt.find(".s-m-header-brand");

					// hide label if brand + label + links > header (in term of width)
					var hideLabel = $$brand.width() + $$label.width() + $$nav.width() + $$links.width() + offset > $$header.width();
					$$label.toggle(hideLabel ? false : true);

					// if label hidden and header still too small, reduce brand fontsize until header has better display
					var sum = $$brand.width() + $$nav.width() + $$links.width();
					var off = 3;
					if (hideLabel == true && (sum > $$header.width())) {
						while (sum > $$header.width()) {
							$$brand.css("font-size", ($$brand.css("font-size").match(/\d+/)[0] - off) + "px");
							off += 3;
							sum = $$brand.width() + $$nav.width() + $$links.width();
						}
					} else {
						$$brand.css("font-size", "2.5em");
					}

				}
			}
			// smartphone
			else {
				// only not on welcome page
				if (!self.controller.state.isWelcomePage) {
					var maxWidth = $$header.width() - ($$nav[0].scrollWidth + $$links[0].scrollWidth + offset);
					$$label.css("max-width", maxWidth);
				}
			}
		},
		/**
		 * Returns HTML for tablet
		 */
		_tabletHtml: function() {
			var self = this;
			var lknCtx = {
				dropdownId: utils.UUID(),
				links: []
			};
			var opts = self.headerOptions;
			var linksCpt = 0;
			for (var p in _linksCtx) {
				if (opts[p] === true) {
					var l = _linksCtx[p];
					l.items.forEach(function(x, idx) {
						x = $.extend({}, x);
						x.label = locale.text(x.label);
						x.ctrlId = self.id;
						x.icon = fontUtils.pageIcon(x.icon);
						if (idx === 0 && lknCtx.links.length > 1) x.divider = l.divider;
						lknCtx.links.push(x);
					});
				}
			}
			// showBack force show back
			var _mainCtx = {
				showBack: (_mustHideSoftwareBackButton() === false) ? (opts.showHome || opts.showBack === true) : false,
				showHome: opts.showHome,
				showBrand: self.controller.state.isWelcomePage,
				labelBrand: opts.labelBrand || "Sage",
				labelProduct: opts.labelProduct || "ERP X3",
				headerLabel: self.controller.prototype.data("$title")
			};
			var htmlUserInfo = _getHtml("tablet", "userInfo", {
				username: globals.getUserName(),
				userrole: globals.getUserRole()
			});
			if (lknCtx.links.length === 0) {
				_mainCtx.userContent = _getHtml("tablet", "infoOnly", {
					userInfo: htmlUserInfo
				});
			} else {
				_mainCtx.userContent = _getHtml("tablet", "dropDown", {
					userInfo: htmlUserInfo,
					dropDownMenu: _getHtml("tablet", "dropDownMenu", lknCtx)
				});
				/** Force overflow visible on s-m-control to display the dropDownMenu**/
				self.$$elmt.css({
					overflow: "visible"
				});
			}
			return _getHtml("tablet", "main", _mainCtx);
		},
		/**
		 * Returns HTML for smartphone
		 */
		_smartPhoneHtml: function(opts, displayController, actionController) {
			var self = this;
			this.resetSmartphonePanels();
			if (opts.showUserpanel) {
				self._smphUserpanel = displayController.panelCreate(_userPanelId, opts);
			}
			var showActions = opts.showActions !== false;
			if (showActions && opts.showActionpanel) {
				self._smphPageactionpanel = displayController.panelCreate(_actionPanelId, opts);
			}
			var backDom = "";
			var labelContent = "";
			var linksContent = "";
			// build backDom only if iPhone
			if (opts.showHome && globals.getLookAndFeel() === "ios") {
				backDom = _getHtml("smartphone", "backDom", {});
			}
			// build label content. 
			labelContent = self._getLabelContent(opts);
			// build links content.
			var hasPanelLinks = self._smphPageactionpanel && self._smphPageactionpanel.hasLinks();
			linksContent = self._getSmartphoneLinksContent(displayController, actionController, hasPanelLinks, showActions);
			var ctx = {
				backDom: backDom,
				labelContent: labelContent,
				linksContent: linksContent,
				panelId: _userPanelId,
				ctrlId: self.id,
				showUserpanel: opts.showUserpanel === true
			};
			return _getHtml("smartphone", "main", ctx);
		},
		/**
		 *	/!\ SMARTPHONE ONLY
		 *	Returns header label content which depends on the current page.
		 *	Label depends whether user is located in home page, app dashboard page, inside an app, or anywhere else
		 */
		_getLabelContent: function(opts) {
			var self = this;
			var labelContent;

			function getContent(isLabelDom, labelVal, opts) {
				if (isLabelDom) {
					return _getHtml("smartphone", "labelDom", {
						label: labelVal
					});
				}
				// brand name
				else {
					return _getHtml("smartphone", "brandDom", {
						labelBrand: opts.labelBrand || "Sage",
						labelProduct: opts.labelProduct || "ERP X3",
					});
				}
			}

			// home page --> brand name
			if (this.controller.state.isWelcomePage) {
				labelContent = getContent(false, null, opts);
			} else {
				// app dashboard page --> app title
				labelContent = getContent(true, self.controller.prototype.data("$title"));
				if (!labelContent) {
					// other cases (for example html static page)
					labelContent = getContent(false, null, opts);
				}
			}
			return labelContent;
		},
		/**
		 *	/!\ SMARTPHONE ONLY
		 *	Returns header links content which depends on the current page.
		 *	- No link on welcome page and application dashboard
		 *	- On query page, links will be "create" and an anchor to open the actions panel (on the right)
		 *	- On detail page, links will be "edit" and an anchor to open the actions panel (on the right)
		 *	- On edit page, links will be "save", "cancel" and an anchor to open the actions panel (on the right)
		 *	Some actions will be displayed in the footer
		 */
		_getSmartphoneLinksContent: function(displayController, actionController, hasPanelLinks, showActions) {
			var linksContent = "";
			this._multiSelManager = multiSelectionMgr.create(actionController, ".s-m-header-nav, .s-m-header-labels, a:not(.s-m-ismultisel)");
			notifications.subscribe(this, ["sm.panel.multiselection.toggle"]);
			this._actionManager = actionManager.create(actionController, "header", {
				addCancelEditAction: "headerLink",
				addMultiSelection: this._multiSelManager && !this.headerOptions.multiSelectionOnly && actionController.isMultiSelectionEnabled(),
				multiSelectionOnly: this.headerOptions.multiSelectionOnly
			});
			var linksArr = this._actionManager.getLinks();
			var cc = [];
			if (hasPanelLinks) {
				cc.push(_getHtml("smartphone", "actionPanelToggle", {
					panelId: _actionPanelId,
					ctrlId: this.id
				}));
			}
			if (showActions && linksArr && linksArr.length > 0) {
				var self = this;
				linksArr.forEach(function(ll, idx) {
					cc.push(_getHtml("smartphone", "actionLink", $.extend(ll, {
						ctrlId: self.id,
						$isHidden: ll.action === "multiSelection"
					})));
				});
			}
			linksContent = cc.join("");
			return linksContent;
		},
		_actHeaderLink: function(uuid) {
			this._actionManager.triggerLink(uuid, this.getActionController()).then(function(success, link, data, options) {
				// Message is displayed 
			}).fail(function(e) {
				if (jsutils.isError(e)) {
					// We shouldn't expect an error except exceptions
					globals.getModal().error("Header action error (smartphone)", e);
				}
			});
		},
		_actHeaderAction: function(id) {
			var panel = id === _actionPanelId ? this._smphPageactionpanel : this._smphUserpanel;
			var side = id === _actionPanelId ? "right" : "left";
			this.controller.panelAttach(side, panel, "toggle");
		},
		_actLinkClicked: function(actionName) { /* triggering action */
			eventListener.triggerAction(this.$$elmt, actionName);
		},

		notifPanelMultiselectionToggle: function(uuid, $$target) {
			this.toggleMultiSelection();
		},

		toggleMultiSelection: function() {
			this._actMultiSelection();
			this.$$elmt.find('[data-action="multiSelection"]').show();
		},

		_actMultiSelection: function(uuid, $$target) {
			var $$hLinks = this.$$elmt.find(".s-m-header-links");
			var enabled = this._multiSelManager.toggle(this, $$hLinks, "dropdown");
			$$hLinks.width(enabled ? "100%" : "initial");
			this.$$elmt.find(".s-m-header-nav, .s-m-header-labels, a:not(.s-m-ismultisel)").toggle(!enabled);
			this.$$elmt.find('[data-action="multiSelection"]').toggle(enabled);

		},

		/**
		 * Page footer can be changed
		 * This method is called each time the header is toggled
		 */
		notifToggled: function(shown) {
			if (shown === false) {
				if (this._multiSelManager) {
					this._multiSelManager.disable();
				}
				return;
			}
		},
		/**
		 * Detach from document
		 */
		detach: function() {
			this.notifToggled(false);
			if (this.$$elmt) {
				this.$$elmt.detach();
			}
		}
	});

exports.Klass = _Klass;
