// syracuse-tablet/html/js/controls/page/ctrlVignetteHeader.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var globals = require('syracuse-tablet/html/js/helpers/globals');
var Base = require('syracuse-tablet/html/js/controls/page/ctrlHeaderBase').Klass;

/**
 * Header displayed whe a vignette is pinned (action triggering is enabled)
 */
var _Klass = utils.defineClass(

	function CtrlVignetteHeader($type, vignette, options) {
		Base.call(this, $type, vignette.controller, options);
		this._vignette = vignette;
	}, Base, {

		destroy: function() {
			Base.prototype.destroy.call(this);
			this._vignette = null;
		},

		buildHtml: function($$parent, controllerDao, buildOptions) {
			Base.prototype.buildHtml.call(this, $$parent, controllerDao, buildOptions);
			// Update drafts menu item - Vignette header is created on demand and has not been notified previouly
			globals.getStorage().draftOperation("draftNotifyChanged");
		},

		getHeaderOptions: function(buildOptions) {
			return {
				showHome: false,
				showAuthoring: false,
				showDevelopment: false,
				showUserpanel: true,
				showActionpanel: false,
				multiSelectionOnly: this.options.multiSelectionOnly === true,
				// To force build
				forceRefresh: true,
				showDrafts: true
			};
		},

		getActionController: function() {
			return this._vignette.getPage();
		},

		/**
		 * Called after the header has been attached
		 * Enable multiselection
		 */
		notifToggled: function(shown) {
			Base.prototype.notifToggled.call(this, shown);
			if (shown === true && this._multiSelManager) {
				this.toggleMultiSelection();
			}
		}

	});

exports.Klass = _Klass;
