// syracuse-tablet/html/js/controls/page/filterSortPanel.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var Base = require('syracuse-tablet/html/js/controls/page/ctrlSidePanel').Klass;
var notifications = require('syracuse-tablet/html/js/helpers/notifications');
var uiUtils = require('syracuse-tablet/html/js/ui/uiUtils');
var globals = require('syracuse-tablet/html/js/helpers/globals');
var fontUtils = require('syracuse-tablet/html/js/ui/fontUtils');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var eventListener = require('syracuse-tablet/html/js/application/eventListener');

var _templates = {
	main: '\
		<nav class="navbar s-m-side-panel {{side}} navbar-inverse ctrl-evt-click" role="navigation">\
			<div class="s-m-scroll-elmt">\
			</div>\
		</nav>'
};

// TODO

var _getHtml = function(name, ctx) {
	var key = name + "compiled";
	var tmpl = _templates[key];
	if (!tmpl) {
		tmpl = _templates[key] = Handlebars.compile(_templates[name]);
	}
	return tmpl(ctx);
};

var _fail = function(e) {
	globals.getModal().error("Error loading applications", e);
};

/**
 * Regular page left panel
 */
var _Klass = utils.defineClass(
	function(controller, parentArray, options) {
		this.parentArray = parentArray;
		Base.call(this, controller, "tablet/x-filtersortpanel", "right", options);
	}, Base, {
		destroy: function() {
			Base.prototype.destroy.call(this);
			this.parentArray = null;
		},
		sstoggle: function(action, callback) {
			var self = this;

			// build panel core if it doesn't exist yet
			if (!self.$$elmt) {
				var $$parent = $(".s-m-panel.s-m-right");
				Base.prototype.buildHtml.call(this, $$parent, null, {}, [this.typeName]);

				// TODO build panel, filter page, sort page, etc
				self.$$panel = $(_getHtml("main", {
					side: "right"
				}));
				self.$$panel.appendTo(self.$$elmt);

				self.$$elmt.show();
			} else {
				//TODO
			}
		},
		_actLinkClicked: function(actionName) {
			/* triggering action */
			eventListener.triggerAction(this.$$elmt, actionName);
		}
	});

exports.Klass = _Klass;
