// syracuse-tablet/html/js/controls/page/regularActionPanel.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var Base = require('syracuse-tablet/html/js/controls/page/ctrlSidePanel').Klass;

var _templates = {
	actionLink: '<li class=""><a href="#" data-action="homeApp" data-params="{{$uuid}}"><i class="{{css}}"></i>&nbsp;<span>{{title}}</span></a></li>'
};

var _actions = '<li class=""><a href="#"><i class="fa fa-1x fa-home"></i> <span>RIGHT REGULAR</span></a></li>';

var _getHtml = function(name, ctx) {
	var key = name + "compiled";
	var tmpl = _templates[key];
	if (!tmpl) {
		tmpl = _templates[key] = Handlebars.compile(_templates[name]);
	}
	return tmpl(ctx);
};

/**
 * Regular page action panel
 */
var _Klass = utils.defineClass(

	function(controller, $type) {
		var self = this;
		Base.call(self, controller, $type, "right", "actions");
	}, Base, {

		buildHtml: function($$parent, controllerDao, buildOptions) {
			var self = this;
			var context = {
				title: "Actions",
				actionsHtml: _actions
			};
			Base.prototype.buildHtml.call(self, $$parent, context);
		},

		afterRender: function() {
			var self = this;
			if (true) {
				self.publishReady();
			}
		}
	});

exports.Klass = _Klass;
