// syracuse-tablet/html/js/controls/page/userSmartphonePanel.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var Base = require('syracuse-tablet/html/js/controls/page/ctrlSidePanel').Klass;
var notifications = require('syracuse-tablet/html/js/helpers/notifications');
var uiUtils = require('syracuse-tablet/html/js/ui/uiUtils');
var globals = require('syracuse-tablet/html/js/helpers/globals');
var fontUtils = require('syracuse-tablet/html/js/ui/fontUtils');

var _templates = {
	navigLink: '<li class="">\
		<a draggable="false"  href="#" data-action="switchApp" data-params="{{appName}}" data-control-id="{{ctrlId}}">\
			<i class="{{icon}}"></i>\
			&nbsp;\
			<span>{{title}}</span>\
		</a>\
	</li>'
};


var _getHtml = function(name, ctx) {
	var key = name + "compiled";
	var tmpl = _templates[key];
	if (!tmpl) {
		tmpl = _templates[key] = Handlebars.compile(_templates[name]);
	}
	return tmpl(ctx);
};

var _fail = function(e) {
	globals.getModal().error("Error loading applications", e);
};

/**
 * Regular page left panel
 */
var _Klass = utils.defineClass(

	function(controller, $type, options) {
		this._$$wait = null;
		Base.call(this, controller, $type, "left", options);
		this.appCount = 0;
	}, Base, {

		destroy: function() {
			Base.prototype.destroy.call(this);
			uiUtils.waitPluginDestroy(this._$$wait);
			this._$$wait = null;
		},

		afterResizeBuildHtml: function(context, pageInfo, orientation, deviceClass) {
			var self = this;
			try {
				var actionsHtml = [],
					curApp = globals.getApplication(),
					title;

				actionsHtml.push(_getHtml("navigLink", {
					title: "title test",
					ctrlId: self.id
				}));


				context.title = title || "User Smartphone";
				context.actionsHtml = actionsHtml.join('');
				Base.prototype.afterResizeBuildHtml.call(self, context, pageInfo, orientation, deviceClass);
				self.publishReady();
			} catch (e) {
				_fail(e);
			}
		}
	});

exports.Klass = _Klass;
