// syracuse-adxadmin/lib/sadfsq/sadfsqFile.js
"use strict";

var helpers = require('syracuse-core/lib/helpers');

exports.SadFsqFile = helpers.defineClass(function(fileDescr) {
	this.filePath = "";
	this.name = "";
	this.extension = "";
	this.adxFolder = "";
	this.adxVol = "";
	this.adxServer = "";

	var idx = fileDescr.indexOf(",");
	this.fileDescriber = fileDescr;
	if (idx !== -1) {
		if (idx === 0) throw new Error("non canonical path TODO");
		this.fileDescriber.split(',').map(function(item, i) {
			switch (i) {
				case 0:
					this.filePath = item;
					break;
				case 1:
					this.name = item;
					break;
				case 2:
					this.extension = item;
					break;
				case 3:
					this.adxFolder = item;
					break;
				case 4:
					this.adxVol = item;
					break;
				case 5:
					this.adxServer = item;
					break;
				default:
					throw new Error("too many part in canonical describer (max 6). TODO");
			}
		});
	} else {
		var lastIdxSep = fileDescr.lastIndexOf('\\');
		if (lastIdxSep === -1) lastIdxSep = fileDescr.lastIndexOf('/');
		if (lastIdxSep !== -1) {
			this.filePath = fileDescr.substring(0, lastIdxSep).replace("\\", "/");
			this.name = fileDescr.substring(lastIdxSep + 1);
			// get extension
			var spl = this.name.split('.');
			this.extension = spl.length > 1 ? spl[1] : null;
			this.name = spl[0];
		}
	}


	this.binary = new Buffer([]); // binary readed or to write (setted during readFile or writeFile)
	this.fileInfo;
	this.err;
	this.fileNum;
}, null, {});
