// syracuse-ui/lib/field/array/record/recordArticle.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Article = require("syracuse-ui/lib/article/article").Article;
var ListContextMenu = require('syracuse-ui/lib/field/array/listContextMenu').ListContextMenu;

function RecordArticle(){
}

exports.RecordArticle = helpers.defineClass(RecordArticle, Article, {
    ensureDataSet: function(){
        return this.dataset = this.articleParent.dataset[this.$serverIndex] || {};
    },
    applyActionLinkChange: function(resources, record){
        Article.prototype.applyActionLinkChange.call(this, resources, record);
        if (resources && (resources.$links !== undefined || resources.$actions !== undefined)) {
            if (!this.contextMenu && !this.$item.$isMenuRecordHidden) {
                (this.contextMenu = new ListContextMenu()).create(this);
            }
            if (resources.$links === null && this.contextMenu) {
                this.contextMenu.clearMenuItems();
            }
        }
    },
    renderRowIndex: function(){
        if (this._rowIndexCell) {
            this._rowIndexCell.textContent = (this.list._store.pagingCapability._$startIndex || 0) + this.$recordIndex;
        }
    },
    applyMetaData: function(metaData){
    },
    dispose: function(){
        delete this.page;
        delete this.list
        delete this.builder;
        if (this.contextMenu) {
            document.controller.disposeObject(this.contextMenu);
            delete this.contextMenu;
        }
        delete this.arrayLevel;
        delete this._rowIndexCell;
        delete this.$recordIndex;
        Article.prototype.dispose.call(this);
    }
});
