// syracuse-ui/lib/field/array/record/recordStore.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var _recordSelectors = {
    multi: require('./selector/multiRecordSelector').MultiRecordSelector,
    single: require('./selector/singleRecordSelector').SingleRecordSelector,
    button: require('./selector/buttonRecordSelector').ButtonRecordSelector,
    row: require('./selector/rowRecordSelector').RowRecordSelector
};

function RecordStore(){

}

exports.RecordStore = helpers.defineClass(RecordStore, null, {
    load: function(list){
        this.list = list;
        this._isAddDisabled = false;
        this._records = [];
        this._map = {};
    },
    hasRecords: function(){
        return this._records && this._records.length > 0;
    },
    loadSelector: function(){
        if (this.selector) {
            document.controller.disposeObject(this.selector);
        }
        var selectorClass = _recordSelectors[this.list.$item.$selectMode];
        if (selectorClass) {
            (this.selector = new selectorClass()).load(this.list);
        }
        else {
            this.selector = {};
        }
    },
    findRecordByServerIndex: function($serverIndex){
        for (var ii = 0, jj = this._records.length; ii < jj; ii++) {
            if (this._records[ii].$serverIndex == $serverIndex) {
                return this._records[ii];
            }
        }
        return null;
    },
    findServerIndex: function(dataRecord){
        for (var ii = 0, jj = this.list.dataset.length; ii < jj; ii++) {
            if (this.list.dataset[ii] == dataRecord) {
                return ii;
            }
        }
        return -1;
    },
    findRecord: function($$selector){
        var $uuid = this.findRecordUiid($$selector)
        return this._map[$uuid];
    },
    findRecordUiid: function($$selector){
        return ($$selector.is("[data-s-record]") ? $$selector : $$selector.closest("[data-s-record]")).attr("data-s-record");
    },
    removeRecords: function(addEmptySlot){
        for (var ii = 0, jj = this._records.length; ii < jj; ii++) {
            var record = this._records[ii];
            delete this._map[record.$uuid];
            this.list.removeRecord(record);
        }
        this._map = {};
        this._records = [];
        if (this.selector && this.selector.clear) {
            this.selector.clear();
        }
        this.list.builder.emptyBody(addEmptySlot);
    },
    setState: function(state){
        if (this._records.length > 0) {
            if (state.$isReadOnly !== undefined || state.$isDisabled != undefined) {
                var metaData = {
                    $isReadOnly: state.$isReadOnly,
                    $isDisabled: state.$isDisabled
                };
                this._records.forEach(function(record, index){
                    state.$isDisabled
                    record.applyMetaData(metaData);
                });
                if (this.list.$capability && state.$isDisabled !== undefined) {
                    this._records.forEach(function(record, index){
                        record.setState(state);
                    });
                }
            }
        }
    },
    /* sort: function(arr, criteria, p){
     this.sortCapability.sort(arr, criteria, p);
     },*/
    appendRecord: function(options){
        var builder = this.list.builder;
        if (builder.isSingleBuilder) {
            options.dataRecord = {
                $singleField: options.dataRecord
            };
            options.$prototype = {
                $properties: {
                    $singleField: this.list.$prototype.$item
                }
            };
        }
        options.dataRecord.$uuid = options.dataRecord.$uuid || options.$recordIndex;
        var record = (this._map[options.dataRecord.$uuid] = new builder.RecordClass());
        record.arrayLevel = "record";
        record.$uuid = options.dataRecord.$uuid;
        record.$serverIndex = options.$serverIndex;
        record.$recordIndex = options.$recordIndex;
        record.$facet = this.list.$recordFacet;
        record.$isEditMode = this.list.$isEditMode;
        record.builder = (record.list = this.list).builder;
        record.$prototype = options.$prototype || this.list.$prototype.$item;
        record.isInsert = options.isInsert;
        if (record.initializeRecord) {
            record.initializeRecord(options);
        }
        this.list.page.initializeNewItem(record, options.$item || {}, this.list);
        record.loadBox(options.dataRecord);
        if (options.isInsert) {
            this._records.splice(options.$recordIndex, 0, record);
        }
        else {
            this._records.push(record);
        }
        return record;
    },
    notifySelectRecords: function(selectedRecords){
        if (selectedRecords) {
            var self = this;
            var sendBag = self.list.page.ensureSendBag(self.list);
            var dataRecords = self.list.ensureDataSet();
            if (self.list.builder.isSingleBuilder) {
                sendBag.dataRecords = dataRecords;
                Object.keys(selectedRecords).forEach(function($uuid){
                    var options = {
                        dataRecord: selectedRecords[$uuid].dataset,
                        $recordIndex: self._records.length
                    };
                    if (self.list.$prototype.$item.$type == "application/x-choice") {
                        options.dataRecord = options.dataRecord.$uuid;
                    }
                    dataRecords.push(options.dataRecord);
                    self.appendRecord(options);
                });
            }
            self.ensureRecordsIndex();
            self.list.page.notifyDataChange(self.list, sendBag.dataRecords);
        }
    },
    ensureRecordsIndex: function(){
        for (var ii = 0, jj = this._records.length; ii < jj; ii++) {
            var record = this._records[ii];
            record.$recordIndex = ii;
            if (record.dataset.$index !== undefined) {
                record.dataset.$index = ii;
            }
            record.$serverIndex = this.findServerIndex(record.dataset);
            record.renderRowIndex();
        }
        this.checkMaxItems();
    },
    doSelectAction: function(){
        var self = this;
        if (self.list.$prototype.$item.$type == "application/x-choice") {
            self.list.builder.loadChoiceSelector();
        }
        else {
            self.list.currentSelectRecords = null;
            self.list.boxParent.openDialog({
                article: self.list,
                $url: self.list.menuItems.$select[0].$url,
                onValidate: function(){
                    self.notifySelectRecords(self.list.currentSelectRecords);
                },
                onSelectRecord: function(selectedRecords){
                    self.list.currentSelectRecords = selectedRecords;
                    return false;
                }
            });
        }
    },
    validate: function(){
        var isValidated = true;
        this._records.forEach(function(record){
            isValidated = record.validate();
        });
        return isValidated;
    },
    disableAdd: function(isAddDisabled){
        this._isAddDisabled = isAddDisabled;
        var metaData;
        if (this.list.$menus) {
            if (this.list.$menus.$create || this.list.$menus.$select) {
                metaData = {
                    $actions: {}
                };
                if (this.list.$menus.$create) {
                    metaData.$actions.$create = {
                        $isDisabled: this._isAddDisabled
                    };
                }
                if (this.list.$menus.$select) {
                    metaData.$actions.$select = {
                        $isDisabled: this._isAddDisabled
                    };
                }
            }
        }
        if (this.list.$capability.insert) {
            (metaData = metaData || {}).$items = [];
            for (var ii = 0, jj = this._records.length; ii < jj; ii++) {
                metaData.$items.push({
                    $actions: {
                        $create: {
                            $isDisabled: this._isAddDisabled
                        }
                    }
                });
            }
        }
        if (metaData) {
            var dataset = {
                $properties: {}
            };
            dataset.$properties[this.list.$item.$bind] = metaData;
            this.list.articleParent.applyChange(dataset);
        }
    },
    checkMaxItems: function(){
        if (this.list.$prototype.$maxItems) {
            if (this._isAddDisabled !== (this.list.$prototype.$maxItems <= this._records.length)) {
                this.disableAdd(!this._isAddDisabled);
            }
        }
    },
    dispose: function(){
        delete this.list;
    }
});
