// syracuse-ui/lib/field/array/record/recordStore.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var PagingCapability = require("./pagingCapability").PagingCapability;
var ReorderCapability = require('./reorderCapability').ReorderCapability;
var SortCapability = require("./sortCapability").SortCapability;
var FilterCapability = require("./filterCapability").FilterCapability;

var _recordSelectors = {
    multi: require('./selector/multiRecordSelector').MultiRecordSelector,
    single: require('./selector/singleRecordSelector').SingleRecordSelector,
    button: require('./selector/buttonRecordSelector').ButtonRecordSelector,
    row: require('./selector/rowRecordSelector').RowRecordSelector
};

function RecordStore(list){
    this.list = list;
    this._isAddDisabled = false;
}

exports.RecordStore = helpers.defineClass(RecordStore, null, {
    load: function(){
        this._records = [];
        this._map = {};
        this.sortCapability = new SortCapability(this.list);
        this.pagingCapability = new PagingCapability(this.list);
        this.filterCapability = new FilterCapability(this.list);
    },
    applySettings: function(newData){
        this.pagingCapability.setOptions(newData);
        this.sortCapability.setOrderBy(newData.$orderBy, newData.$startLetter);
        this.filterCapability.setWhere(newData.$where);
    },
    loadSelector: function(){
        if (this.selector) {
            document.controller.disposeObject(this.selector);
        }
        var selectorClass = _recordSelectors[this.list.$item.$selectMode];
        if (selectorClass) {
            (this.selector = new selectorClass()).load(this.list);
        }
        else {
            this.selector = {};
        }
    },
    findRecord: function($$selector){
        var $uuid = this.findRecordUiid($$selector)
        return this._map[$uuid];
    },
    findRecordUiid: function($$selector){
        return ($$selector.is("[data-s-record]") ? $$selector : $$selector.closest("[data-s-record]")).attr("data-s-record");
    },
    removeRecords: function(addEmptySlot, redraw){
        var self = this;
        self._records.forEach(function(recordArticle){
            delete self._map[recordArticle.$uuid];
            self.list.removeItem(recordArticle);
        });
        self._map = {};
        self._records = [];
        if (self.selector && self.selector.clear) {
            self.selector.clear();
        }
        self.list.builder.emptyBody(addEmptySlot, redraw);
    },
    setState: function(state){
        if (this._records.length > 0) {
            if (state.$isReadOnly !== undefined || state.$isDisabled != undefined) {
                var metaData = {
                    $isReadOnly: state.$isReadOnly,
                    $isDisabled: state.$isDisabled
                };
                this._records.forEach(function(record, index){
                    state.$isDisabled
                    record.applyMetaData(metaData);
                });
                if (this.list.$capability && state.$isDisabled !== undefined) {
                    this._records.forEach(function(record, index){
                        record.setState(state);
                    });
                }
            }
        }
    },
    setSortCapability: function(){
        this.sortCapability.load();
    },
    setReorderCapability: function(reorder){
        if (reorder) {
            this.list.$item.$isRowIndexVisible = true;
            if (!this.reorderCapability) {
                (this.reorderCapability = new ReorderCapability(this.list)).setDraggable();
            }
        }
        else {
            if (this.reorderCapability) {
                document.controller.disposeObject(this.reorderCapability);
                delete this.reorderCapability;
            }
        }
    },
    onReorder: function(sourceRecord, targetRecord, $insert){
        var newRecords = [];
        this._records.splice(sourceRecord.$recordIndex, 1);
        this._records.forEach(function(record){
            if (record == targetRecord) {
                if ($insert == "insertAfter") {
                    record.$recordIndex = newRecords.length;
                    newRecords.push(record);
                    sourceRecord.$recordIndex = newRecords.length;
                    newRecords.push(sourceRecord);
                }
                else {
                    sourceRecord.$recordIndex = newRecords.length;
                    newRecords.push(sourceRecord);
                    record.$recordIndex = newRecords.length;
                    newRecords.push(record);
                }
            }
            else {
                record.$recordIndex = newRecords.length;
                newRecords.push(record);
            }
        });
        this._records = newRecords;
        sourceRecord.notifyReorder(targetRecord, $insert);
    },
    sort: function(arr, criteria, p){
        this.sortCapability.sort(arr, criteria, p);
    },
    appendRecord: function(options){
        var builder = this.list.builder;
        if (builder.isFusionTemp && options.isInsert) {
            options.$recordIndex = -options.$recordIndex - 1;
        }
        if (builder.isSingleBuilder) {
            options.dataRecord = {
                $singleField: options.dataRecord
            };
            options.$prototype = {
                $properties: {
                    $singleField: this.list.$prototype.$item
                }
            };
        }
        options.dataRecord.$uuid = options.dataRecord.$uuid || options.$recordIndex;
        var record = (this._map[options.dataRecord.$uuid] = new builder.RecordClass());
        record.$layoutOptions = builder.$layoutOptions;
        record.$recordUuid = options.dataRecord.$uuid;
        record.$recordIndex = options.$recordIndex;
        record.$facet = this.list.$recordFacet;
        record.$isEditMode = this.list.$isEditMode;
        record.builder = (record.list = this.list).builder;
        record.$prototype = options.$prototype || this.list.$prototype.$item;
        record.isInsert = options.isInsert;
        if (record.initializeRecord) {
            record.initializeRecord(options);
        }
        document.itemFactory.initializeItem(record, options.$item || {}, this.list);
        record.loadBox(options.dataRecord);
        if (options.isInsert) {
            this._records.splice(options.$recordIndex, 0, record);
        }
        else {
            this._records.push(record);
        }
        return record;
    },
    setRecordsIndex: function(drawRowIndex){
        this._records.forEach(function(record, index){
            record.dataset.$index = record.$recordIndex = index;
            if (drawRowIndex) {
                //  debugger;
            }
        });
        this.checkMaxItems();
    },
    checkMaxItems: function(){
        if (this.list.$prototype.$maxItems) {
            var self = this;
            if (self._isAddDisabled !== (self.list.$prototype.$maxItems <= self._records.length)) {
                self._isAddDisabled = !self._isAddDisabled;
                var metaData;
                if (self.list.$menus) {
                    var $actions;
                    ["$create", "$select"].forEach(function($bind){
                        if (self.list.$menus[$bind]) {
                            ($actions = $actions || {})[$bind] = {
                                $isDisabled: self._isAddDisabled
                            };
                        }
                    });
                    if ($actions) {
                        metaData = {
                            $actions: $actions
                        };
                        
                    }
                }
                if (self.list.$capability.insert) {
                    (metaData = metaData || {}).$items = self._records.map(function(){
                        return {
                            $actions: {
                                $create: {
                                    $isDisabled: self._isAddDisabled
                                }
                            }
                        };
                    });
                }
                if (metaData) {
                    var dataset = {
                        $properties: {}
                    };
                    dataset.$properties[this.list.$item.$bind] = metaData;
                    this.list.getArticleParent().applyChange(dataset);
                }
            }
        }
    },
    dispose: function(){
        delete this.list;
    }
});
