// syracuse-ui/lib/desktop/field/vignette/htmlVignetteBuilder.js
"use strict";
var HtmlVignetteAuthoring = require("./htmlVignetteAuthoring").HtmlVignetteAuthoring;
exports.getSettings = function(field){

};
exports.load = function(field){
    var url = "http://www.lemonde.fr/" || field.$field.$location.$url;
    var url = field.$field.$location.$url;
    var title = field.$$title.text();
    var link = document.createElement("a");
    link.className = field.$skin + "-title-link";
    link.setAttribute("href", url);
    link.setAttribute("target", "_blank");
    link.setAttribute("title", field.$field.$location.$title);
    $(link).appendTo(field.$$title.empty()).text(title);
    var iframe = document.createElement("iframe");
    iframe.setAttribute("src", url);
    iframe.style.width = field.$item.$width || "100%";
    iframe.style.height = field.$item.$height || "100%";
    field.$$body.empty().append(field.$$iframe = $(iframe));
};

exports.applyDesignMetaData = function(field, metadata, onAuthoring){
    if (metadata.$url) {
        if (field.$$iframe) {
            field.$$iframe[0].setAttribute("src", url);
        }
    }
    if (metadata.$width) {
        field.$item.$width = metadata.$width;
        if (field.$$iframe) {
            field.$$iframe[0].style.width = field.$item.$width;
        }
    }
    if (metadata.$height) {
        field.$item.$height = metadata.$height;
        if (field.$$iframe) {
            field.$$iframe[0].style.height = field.$item.$height;
        }
    }
};
exports.maximize = function(field){
    exports.onWindowResize(field);
    document.site.addResizeListener(field);
};
exports.minimize = function(field){
    field.$$iframe[0].style.height = field.$item.$height || "100%";
    document.site.removeResizeListener(field);
};
exports.onWindowResize = function(field){
    var body = document.site.$$body;
    var maxHeight = document.site.$$body.height();
    var iframe = field.$$iframe[0];
    iframe.style.height = maxHeight + "px";
    iframe.style.height = (maxHeight - (body[0].scrollHeight - maxHeight)) + "px";
};
exports.getAuthoringWidget = function(){
    return new HtmlVignetteAuthoring();
};

exports.dispose = function(field){
    document.site.removeResizeListener(field);
};
