// syracuse-ui/lib/desktop/field/vignette/htmlVignetteAuthoring.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Article = require("syracuse-ui/lib/common/article/article").Article;

function HtmlVignetteAuthoring(){

}

exports.HtmlVignetteAuthoring = helpers.defineClass(HtmlVignetteAuthoring, Article, {
    onNotifyRecordChange: function(value, binding){
        var metaData = {};
        metaData[binding] = value;
        /*if (binding == "$format") {
         this.applyChange(metaData);
         }*/
        this.designedField.applyDesignMetaData(metaData, true);
		document.site.authoringView.notifyUpdate();
        return null; //cancel notify
    },
    applyChange: function(newData){
        if (newData) {
        
        }
        Article.prototype.applyChange.call(this, newData);
        
    },
    loadBox: function(initData){
        this.$prototype = {
            "$properties": {
                $locationTitle: {
                    $type: "application/x-string",
                    $title: "Title"
                },
                $locationUrl: {
                    $type: "application/x-string",
                    $title: "url"
                },
                $width: {
                    $type: "application/x-string",
                    $title: "Width"
                },
                $height: {
                    $type: "application/x-string",
                    $title: "Height"
                }
            },
            $width: "100%",
            $height: "100%",
            $locationTitle: this.designedField.$field.$location.$title,
            $locationUrl: this.designedField.$field.$location.$url,
        };
        this.$item = {
            $category: "section",
            $title: "HTML Vignette",
            $layout: {
                $items: [{
                    $bind: "$locationTitle",
                    $isEditMode: true,
                    $labelWidth: "auto",
                    $skin: "s-author-field"
                }, {
                    $bind: "$locationUrl",
                    $isEditMode: true,
                    $labelWidth: "auto",
                    $skin: "s-author-field"
                }, {
                    $layout: {
                        $layoutType: "side",
                        $items: [{
                            $bind: "$width",
                            $isEditMode: true,
                            $labelWidth: "auto",
                            $skin: "s-author-field"
                        }, {
                            $bind: "$height",
                            $isEditMode: true,
                            $labelWidth: "auto",
                            $skin: "s-author-field"
                        }]
                    }
                }]
            }
        };
        Article.prototype.loadBox.call(this, initData);
    }
});
