// @sage/syracuse-lib/src/collaboration/advancedScripts/helpers.ts
"use strict";

export function loadMenusMap(_, db, factoryFilter, filter) {
	var map = {};
	db.fetchInstances(_, db.getEntity(_, "menuItem"), {
		sdataWhere: (filter ? "(" + filter + ") and (" + factoryFilter + ")" : factoryFilter),
		rawResults: "unescaped"
	}).forEach(function(dd) {
		map[dd._id] = dd;
	});
	return map;
};

export function loadBlocksMap(_, db, factoryFilter, filter) {
	var map = {};
	db.fetchInstances(_, db.getEntity(_, "menuBlock"), {
		sdataWhere: (filter ? "(" + filter + ") and (" + factoryFilter + ")" : factoryFilter),
		rawResults: "unescaped"
	}).forEach(function(dd) {
		map[dd._id] = dd;
	});
	return map;
};

export function loadModulesMap(_, db, factoryFilter, filter) {
	var map = {};
	db.fetchInstances(_, db.getEntity(_, "menuModule"), {
		sdataWhere: (filter ? "(" + filter + ") and (" + factoryFilter + ")" : factoryFilter),
		rawResults: "unescaped"
	}).forEach(function(dd) {
		map[dd._id] = dd;
	});
	return map;
};

export function loadNavPages(_, db, factoryFilter, filter) {
	return db.fetchInstances(_, db.getEntity(_, "navigationPage"), {
		sdataWhere: (filter ? "(" + filter + ") and (" + factoryFilter + ")" : factoryFilter),
		rawResults: "unescaped"
	});
};
