// @sage/syracuse-react/src/profile/endpoint.ts
"use strict";
import * as Sdata from '../dataStore/sdata';

export interface IEndpointKey {
    application: string,
    contract: string,
    dataset: string
}
export interface IEndpoint extends IEndpointKey, Sdata.IEntity {
    description: string,
    nature: string
};

export const administration: IEndpointKey = {
    application: "syracuse",
    contract: "collaboration",
    dataset: "syracuse"
};
export const search: IEndpointKey = {
    application: "syracuse",
    contract: "search",
    dataset: "syracuse"
};
export const x3js: IEndpointKey = {
    application: "x3js",
    contract: "studio",
    dataset: ""
};

export function toUrlString(key: IEndpointKey) {
    return key && (key.application + "/" + key.contract + "/" + key.dataset);
}
export function fromUrlString(url: string):IEndpointKey {
    let parts = url.split("/");
    return {
        application: parts[0],
        contract: parts[1],
        dataset: parts[2]
    };
}
export function toString(key: IEndpointKey) {
    return key && (key.application + "." + key.contract + "." + key.dataset);
}
export function matchContract(key1: IEndpointKey, key2: IEndpointKey) {
    if (key1 && key2) {
        return key1.application == key2.application && key1.contract == key2.contract;
    }
}

export function cloneKey(key: IEndpointKey){
    return {
         application: key.application,
        contract: key.contract,
        dataset: key.dataset
    };
}
