// @sage/syracuse-react/src/profile/profileController.tsx
"use strict";
import * as React from 'react';
import * as Sdata from '../dataStore/sdata';
import { culture, ILocalePreferences } from '../culture/culture';
import * as Cookie from '../utility/storage/cookie';
import { applyTheme, ITheme } from '../style/styleSheet';
import { Profile, IRole } from './profile';
import * as Actions from './actions';
import { IEndpoint } from './endpoint';
import { Ajax } from '../ajax/ajax';
import { route } from '../route/route';
import { syra_site } from '../syra/utility/syra_site';
import { syra_help } from '../syra/utility/syra_help';
import { ProfileBar } from './bar/profileBar';
import { OfficeController } from '../office/officeController';
import { OfficeProfileBar } from '../office/officeProfileBar';
import * as BookmarkController from '../navigation/bookmark/bookmarkController';
import AppConnController from './applicationConnection/appConnController';
import { IProfileBookmark } from './bar/profileBarProps';
import { showDiagnoses } from '../alert/alert';
import { getLocalString } from '../culture/localString';
import { syra_fusion } from '../syra/utility/syra_fusion';
import * as clientSettings from '../utility/clientSettings';
import * as Pendo from '../utility/vendor/pendo';

interface IProps extends IProfileBookmark {
    office: OfficeController,
    searchFacet: boolean,
    classicMode: boolean,
    profile: Profile
    onLogout: () => void,
    onReady: (profile: Profile) => void,
    onToggleSitemap: (triggerToggle: () => void) => void,
    onSwitchApplicationMode: (event: React.MouseEvent<HTMLElement>) => void
}

export class ProfileController extends React.Component<IProps, {
    searchFieldVisible: boolean
}>{
    appConnController: AppConnController;
    applicationConnectionUrl: string;

    constructor(props: IProps, context: any) {
        super(props, context);
        this.state = {
            searchFieldVisible: false
        };
        Ajax.setup(culture.languageCode);
        Actions.loadDefaultLocalStrings();
    }
    get sessionId(): string {
        return Cookie.get("syracuse.sid." + window.location.port) || "";
    }
    getNavigationContext(profile: Profile, done: () => void) {
        profile.bookmarks = profile.sitemap = null;
        Actions.dispose();
        BookmarkController.fetch(profile, (diagnoses?: Sdata.IDiagnose[]) => {
            Actions.getSitemap(profile);
            diagnoses && showDiagnoses(diagnoses, getLocalString("profile_bookmarks_fetch_failed"));
            done();
        });
    }
    onToggleRolesPopup = (opened: boolean) => {
        if (opened && !this.props.profile.roles) {
            Actions.getRoles(this.props.profile, (roles: IRole[]) => {
                this.props.profile.roles = roles;
                this.setState({});
            });
        }
    }
    onToggleEndpointsPopup = (opened: boolean) => {
        if (opened && !this.props.profile.endpoints) {
            Actions.getEndpoints(this.props.profile, (endpopints: IEndpoint[]) => {
                this.props.profile.endpoints = endpopints;
                this.setState({});
            });
        }
    }
    componentDidMount() {
        Actions.getProfile((profile: Profile) => {
            profile.saveRestorePoint();
            Actions.changeContext(profile);
            syra_site.userProfile = profile;
            //to improve for moment used only by ui-text
            (window as any).syra_context.developpementMode = profile.developpementMode;
            clientSettings.traverseClientSettings(profile.clientSettings, { "userProfile": profile.dataset });
            Actions.loadLocalStrings((ok: boolean) => {
                applyTheme(profile.selectedTheme);
                profile.clientSettings.pendo && Pendo.loadPendo(profile.clientSettings.pendo);
                this.getNavigationContext(profile, () => this.props.onReady(profile));
                //compatibility with v11. Used by cti extension
                syra_site.showExtensions(profile);
            });
        });
    }
    componentDidUpdate() {
        if (this.props.profile) {
            let curAppConnUrl = this.props.profile.applicationConnectionUrl;
            if (curAppConnUrl != this.applicationConnectionUrl) {
                this.appConnController && this.appConnController.dispose();
                this.appConnController = null;
                this.applicationConnectionUrl = curAppConnUrl;
                if (curAppConnUrl) {
                    this.appConnController = new AppConnController();
                    this.appConnController.load(curAppConnUrl, (appConn) => {
                        if (appConn) { //null if error
                            appConn.applyDelta(this.props.profile.applicationConnectionData);
                        }
                        this.forceUpdate();
                    });
                }
                else {
                    this.setState({});
                }
            }
        }
    }
    onSelectTheme = (theme: ITheme) => {
        Actions.changeTheme(this.props.profile, theme, () => route.renderCurrentUrl());
    }
    onSelectLocale = (locale: ILocalePreferences) => {
        Actions.changeLocale(this.props.profile, locale, route.reload);
    }
    onSelectRole = (role: IRole) => {
        Actions.changeRole(this.props.profile, role, () => route.renderCurrentUrl());
    }
    onSelectEndpoint = (endpoint: IEndpoint) => {
        Actions.changeEndpoint(this.props.profile, endpoint, () => {
            this.getNavigationContext(this.props.profile, () => route.gotoHome(true)); //force page replacement
        });
    }
    onEditAppConn = () => {
        Actions.changeApplicationConnection(this.props.profile, this.appConnController, () => route.reload());
    }
    onGoToLegal = () => {
        syra_help.openLegalPage();
    }
    onGoToUser = () => {
        route.push(this.props.profile.userUrl);
    }
    onUploadPhoto = () => {
        alert("To do");
    }
    onGotoHelpCenter = () => {
        route.openNewWindow("/help/" + culture.languageCode + "/index.htm?format=text/html$ep=" + this.props.profile.selectedEndpoinUuid);
    }
    onGotoHelpShortcuts = () => {
        route.openNewWindow("/help/" + culture.languageCode + "/KEYS/index.htm?format=text/html");
    }
    onToggleSearch = () => {
        if (this.props.classicMode) {
            syra_fusion.runGoTo();
        }
        else {
            this.setState({
                searchFieldVisible: !this.state.searchFieldVisible
            });
        }
    }
    componentWillUnmount() {
        this.appConnController && this.appConnController.dispose();
        Actions.dispose();
    }
    renderOfficeBar() {
        return <OfficeProfileBar profile={this.props.profile} />;
    }
    renderDefaultBar() {
        return <ProfileBar
            onHeaderClick={this.props.onSwitchApplicationMode}
            onGoHome={route.gotoHome}
            onToggleSitemap={this.props.onToggleSitemap}
            onGotoHelpCenter={this.onGotoHelpCenter}
            onGotoHelpShortcuts={this.onGotoHelpShortcuts}
            onToggleRolesPopup={this.onToggleRolesPopup}
            onSelectRole={this.onSelectRole}
            onToggleEndpointsPopup={this.onToggleEndpointsPopup}
            onSelectEndpoint={this.onSelectEndpoint}
            onSelectLocale={this.onSelectLocale}
            onSelectTheme={this.onSelectTheme}
            onGoToLegal={this.onGoToLegal}
            onGoToUser={this.onGoToUser}
            searchFieldVisible={!this.props.classicMode && (this.props.searchFacet || this.state.searchFieldVisible)}
            onToggleSearch={this.onToggleSearch}
            onUploadPhoto={this.onUploadPhoto}
            appConnFields={this.appConnController && this.appConnController.fields}
            onEditAppConn={this.onEditAppConn}
            classicMode={this.props.classicMode}
            profile={this.props.profile}
            onLogout={this.props.onLogout}
            bookmarkDisabled={this.props.bookmarkDisabled}
            mainPageBookmark={this.props.mainPageBookmark}
            onBookmarked={this.props.onBookmarked}
            onBookmarkClick={this.props.onBookmarkClick}
        />;
    }
    render() {
        if (this.props.profile) {
            return this.props.office ? this.renderOfficeBar() : this.renderDefaultBar();
        }
        return null;
    }
}
