// syracuse-ui/lib/field/alphaNumField.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Field = require('syracuse-ui/lib/field/field').Field;


function PhoneField(){
}

exports.PhoneField = helpers.defineClass(PhoneField, Field, {
    initialize: function(){
        this.$isDetailLinkIgnore = true;
    },
    getHtmlInput: function(){
        return "<input type='tel'/>";
    },
    setDataValue: function(value, record){
        this.currentValue = value;
        if (this.$isEditMode) {
            this.$$input.val(this.currentValue);
        }
        else {
            this.$$fieldValue.empty().append($("<a class='s-phone-link'/>").attr("href", "callto:" + value).text(value));
        }
    }
});

function UrlField(){
}

exports.UrlField = helpers.defineClass(UrlField, Field, {
    initialize: function(){
        this.$isDetailLinkIgnore = true;
    },
    getHtmlInput: function(){
        return "<input type='url'/>";
    },
    setDataValue: function(value, record){
        this.currentValue = value;
        if (this.$isEditMode) {
            this.$$input.val(this.currentValue);
        }
        else {
            this.$$fieldValue.empty().append($("<a class='s-field-menus-link'/>").attr("href", value).text(value));
        }
    }
});

function EmailField(){
}

exports.EmailField = helpers.defineClass(EmailField, Field, {
    initialize: function(){
        this.$isDetailLinkIgnore = true;
    },
    validateType: function(errors, value){
        var atpos = value.indexOf("@");
        var dotpos = value.lastIndexOf(".");
        if (atpos < 1 || dotpos < atpos + 2 || dotpos + 2 >= value.length) {
            errors.push(this.getLocalize().email_invalid);
        }
    },
    getHtmlInput: function(){
        return "<input type='email'/>";
    },
    setDataValue: function(value, record){
        this.currentValue = value;
        if (this.$isEditMode) {
            this.$$input.val(this.currentValue);
        }
        else {
            this.$$fieldValue.empty().append($("<a rel='nofollow' class='s-email-link'/>").attr("href", "mailto:" + value).text(value));
        }
    }
});
