// syracuse-ui/lib/field/discussionFeedField.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Field = require('syracuse-ui/lib/field/field').Field;

function DiscussionFeedField() {}

exports.DiscussionFeedField = helpers.defineClass(DiscussionFeedField, Field, {
	loadBox: function() {
		Field.prototype.loadBox.call(this);
		var tarea = document.createElement('textarea');
		this.tarea = tarea;
		this.domItem.appendChild(tarea);
		var save = document.createElement('a');
		save.onclick = function() {
			$.ajax({
				type: "GET",
				url: "/sdata/syracuse/collaboration/syracuse/endPoints('c0a3cd23-a3c4-4e0b-9900-c055c864e39d')/discussionFeed?message=" + tarea.value,
				success: function(data, statusText, xhr) {
					alert("SUCCESS");
				},
				error: function(xhr, message, ex) {
					alert("ERROR: " + message);
				}
			});
		};
		save.setAttribute('href', "dummy");
		save.textContent = "Save";
		this.domItem.appendChild(save);
	},
	setDataValue: function(value) {
		this.tarea.value = value && value.message;
	},
});
