// syracuse-ui/lib/field/documentField.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Field = require('syracuse-ui/lib/field/field').Field;

function _setImageAnchorDefaultSize(field) {
	field.anchor.style.height = field.$item.$imageHeight;
	if (field.$item.$inplace) {
		field.anchor.style.width = field.$item.$imageWidth;
	}
}

function _setImageFile(field, file) {
	// if value != null, get data url from prototype
	var value = field.currentValue;
	if (file) {
		field._file = file;
		_showImageFile(field);
	} else {
		_removeImageArea(field);
		if (field._imgLoader) {
			field._imgLoader.onload = field._imgLoader.onerror = field._imgLoader = null;
		}
		if (field.anchor) {
			syra_site.dom.empty(field.anchor);
			field.anchor.setAttribute("href", "#");
			field.anchor.setAttribute("disabled", true);
			_setImageAnchorDefaultSize(field);
		}
		_showBtn(field, field.removeBtn, false);
	}
}

function _saltUrl(url) {
	if (!url)
		return null;
	var language = syra_site.localizer.getLanguage();
	var sep = (url.indexOf("?") >= 0) ? "&" : "?";
	return url + sep + "salt=" + Math.round(Math.random() * 1000) + (language ? "&language=" + language : "");
}

var _imageLoader = {
	add: function(field) {
		field._isImageLoaded = false;
		_imageLoader.clear(field);
		field._imgLoader = document.createElement("img");
		field._imgLoader.onload = function() {
			if (!this.syraIsClear) {
				// handling image display waiting icon
				clearTimeout(field._waitingTimeOut);
				if (field.anchor && field._imgLoader) {
					field._isImageLoaded = true;
					_imageLoader.clearWait(field);
					_addImageArea(field);
					field.resizeField();
					field._imageArea.style.backgroundSize = "contain";
					field._imageArea.style.backgroundImage = "";
					field._imageArea.style.backgroundImage = "url(\"" + field.imageUrl + "\")";
				}
			}
		};
		field._imgLoader.onerror = function() {
			if (!this.syraIsClear) {
				clearTimeout(field._waitingTimeOut);
				field._isImageLoaded = true;
				if (field.anchor && field._imgLoader) {
					_imageLoader.clearWait(field);
					if (field._imageArea) {
						field._imageArea.style.backgroundImage = "none";
					}
				}
			}
		};
		// handler to display waiting icon
		field._waitingTimeOut = setTimeout(function() {
			if (field.anchor && !field._imageWait && !field._isImageLoaded) {
				field._imageWait = document.createElement("div");
				field._imageWait.className = "s-field-img-img-wait";
				field.anchor.appendChild(field._imageWait);
			}
		}, 500);
		field._imgLoader.src = field.imageUrl;
	},
	clearWait: function(field) {
		if (field._imageWait && field._imageWait.parentNode) {
			field._imageWait.parentNode.removeChild(field._imageWait);
			delete field._imageWait;
		}
	},
	clear: function(field) {
		_imageLoader.clearWait(field);
		if (field._imgLoader) {
			field._imgLoader.syraIsClear = true;
			field._imgLoader.onload = field._imgLoader.onerror = field._imgLoader = null;
		}
	}
};

function _removeImageArea(field) {
	if (field._imageArea) {
		if (field._imageArea.parentNode) {
			field._imageArea.parentNode.removeChild(field._imageArea);
		}
		delete field._imageArea;
	}
}

function _addImageArea(field) {
	_removeImageArea(field);
	var img = field._imageArea = document.createElement("div");
	img.className = field.$skin + "-img-img";
	img.style.width = field.$item.$imageWidth;
	img.style.height = field.$item.$imageHeight;
	img.style.backgroundSize = "contain";
	img.style.backgroundRepeat = "no-repeat";
	field.anchor.appendChild(img);
}

function _showImageFile(field) {
	var value = field.currentValue;
	field.imageUrl = (value && value.$url) || field.$dataUrl;
	if (field.imageUrl) {
		field.imageUrl = syra_site.expressionMaker.parse(field.articleParent, field.imageUrl);
		// key value encoding
		var u = field.imageUrl.split("?");
		var p = u[0].split("/");
		p.push(encodeURIComponent(p.pop()));
		field.imageUrl = _saltUrl([p.join("/"), (u[1] ? u[1] : "")].join("?"));

		if (!field._imageArea) {
			field._dataValue.style.display = '';
		}
		field.anchor.style.height = "";
		_showBtn(field, field.removeBtn, true);
		if (field.$item.$isAutoSize) {
			if (!field.$item.$inplace) {
				field.anchor.style.minWidth = "50px";
			}
			_imageLoader.add(field);
		} else {
			_addImageArea(field);
			field._imageArea.style.backgroundImage = "";
			field._imageArea.style.backgroundImage = "url(\"" + field.imageUrl + "\")";
		}
		if (field.anchor) {
			field.anchor.setAttribute("target", "_blank");
			field.anchor.setAttribute("href", field.imageUrl);
			field.anchor.setAttribute("disabled", false);
		}
	}
}

function _setFile(field, file) {
	if (file) {
		var $type = file.$contentType || file.type || field.$field.$contentType || "";
		field.$fileUrl = field.currentValue && field.currentValue.$url ? field.currentValue.$url : null;
		field.$fileUrl = syra_site.expressionMaker.parse(field.articleParent, field.$fileUrl || field.$dataUrl);
		if (field.$item.$isEmbedded) {
			var slot = document.createElement("div");
			slot.className = field.$skin + "-embed-slot";
			slot.setAttribute("width", field.$item.$imageWidth || "100%");
			slot.setAttribute("height", field.$item.$imageHeight || "100%");
			if (field.isPDFFile()) {
				field.$fileUrl += "#view=FitH&pagemode=thumbs&search=pdfobject&toolbar=0&statusbar=0&messages=0&navpanes=1";
			}
			slot.innerHTML = '<object width="100%" height="100%" type="' + $type + '" data="' + field.$fileUrl + '"  ></object>';
			field._dataValue.insertBefore(slot, field._dataValue.firstChild);
		} else {
			if (!field._imgLoader) {
				field._imgLoader = document.createElement("img");
				field._imgLoader.style.display = "none";
				field._imgLoader.onload = function() {
					clearTimeout(field._defaultFileTimeOut);
					field._imgLoader.className = field.$skin + "-file-img";
					field._imgLoader.style.display = "";
				};
				field.anchor.appendChild(field._imgLoader);
			}
			field._defaultFileTimeOut = setTimeout(function() {
				if (field.anchor) {
					syra_site.dom.toggleClass(field.anchor, field.$skin + "-file-default", true);
				}
			}, 1000);

			var image = syra_site.siteFunctions.getDocumentIcon($type);
			field._imgLoader.src = syra_site.$item.$iconPath + "/file/s-file-" + (image || "default") + ".png";
			_showBtn(field, field.removeBtn, true);

			if (field.anchor) {
				field.anchor.setAttribute("target", "_blank");
				field.anchor.setAttribute("href", field.$fileUrl);
				field.anchor.setAttribute("disabled", false);
			}
		}

	} else {
		delete field.$fileUrl;
		if (field._imgLoader) {
			delete field._imgLoader.onload;
			delete field._imgLoader;
		}
		if (field.anchor) {
			syra_site.dom.empty(field.anchor);
			syra_site.dom.toggleClass(field.anchor, field.$skin + "-file-default", false);
			field.anchor.setAttribute("href", "#");
			field.anchor.setAttribute("disabled", true);
		}
		_showBtn(field, field.removeBtn, false);
	}
}

function _validateFile(field, file) {
	var errors = [];
	if (file.type !== undefined) {
		if (field.$field.$acceptedTypes) {
			if (field.$field.$acceptedTypes.indexOf(file.type) < 0) {
				//check office types
				if (!_matchOfficeType(file.type, field.$field.$acceptedTypes)) {
					errors.push(syra_local.fdocNotSupported);
				}
			}
		}
	}
	if (field.$field.$maxByteLength && file.size > field.$field.$maxByteLength) {
		errors.push(syra_local.fdocTooLarge);
	}
	if (errors.length > 0) {
		field.showErrors(errors);
		return false;
	}
	return true;
}

function _onBeginUpload(field, file) {
	field._progress = document.createElement("div");
	field._progress.className = "s-file-progress";
	field._info.appendChild(field._progress);
	field.progressValue = document.createElement("div");
	field.progressValue.className = "s-file-progress-value";
	field._progress.appendChild(field.progressValue);
	_disableButtons(field, true);
	field._legend.textContent = syra_local.fdocUpload + ":" + " " + file.name + " 0%";
}

function _onProgressUpload(field, file, progress) {
	field._legend.textContent = syra_local.fdocUpload + ":" + " " + file.name + " " + progress + "%";
	field.progressValue.style.width = progress + "%";
}

function _appendActions(field, $skin) {
	var btns = document.createElement("div");
	btns.className = $skin + "-doc-btns";

	field.inputFile = document.createElement("input");

	syra_site.setSpecificAttributes(field.inputFile);

	field.inputFile.setAttribute("type", "file");
	if (field.acceptFilter) {
		field.inputFile.setAttribute("accept", field.acceptFilter);
	}
	if (field.$item.$enableMultiple) {
		field.inputFile.setAttribute("multiple", true);
	}
	//btns.appendChild(field.inputFile);
	field._bindInputFileChange(true);

	btns.appendChild(field.inputFile);
	field.inputFile.className = "s-field-input-uploader";

	var addButton = field.$item.$inplace ? syra_menus.addIconButton : syra_menus.addTextButton;

	field.selectBtn = addButton(syra_local.fdocSelectFile, $skin + "-doc-btn s-field-doc-upload", "onSelectClick", null, "upload");
	btns.appendChild(field.selectBtn);

	field.removeBtn = addButton(syra_local.fdocRemoveFile, $skin + "-doc-btn s-field-doc-remove", "onRemoveClick", null, "cancel");
	_showBtn(field, field.removeBtn, false);
	btns.appendChild(field.removeBtn);

	field._right.appendChild(btns);
}

function _disableButtons(field, isDisabled) {
	field.selectBtn && syra_site.dom.disableItem(field.selectBtn, isDisabled);
	field.removeBtn && syra_site.dom.disableItem(field.removeBtn, isDisabled);
}

function _showBtn(field, btn, show) {
	if (btn) {
		if (field.$item.$inplace) {
			btn.style.visibility = (show && !field.$isReadOnly) ? "visible" : "hidden";
		} else {
			btn.style.display = (show && !field.$isReadOnly) ? "" : "none";
		}
	}
}

function _matchGroupType(type, group) {
	for (var ii = 0, jj = group.length; ii < jj; ii++) {
		if (type.indexOf(group[ii]) >= 0) {
			return true;
		}
	}
	return false;
}

function _matchOfficeType($type, $acceptedTypes) {
	var groups = [
		["msword", "officedocument.word"],
		["ms-excel", "officedocument.spreadsheet"],
		["ms-powerpoint", "officedocument.presentation"]
	];
	for (var ii = 0, jj = groups.length; ii < jj; ii++) {
		var group = groups[ii];
		if (_matchGroupType($type, group)) {
			for (var mm = 0, kk = $acceptedTypes.length; mm < kk; mm++) {
				if (_matchGroupType($acceptedTypes[mm], group)) {
					return true;
				}
			}
		}
	}
	return false;
}


function _getFileType(mediaType) {
	if (mediaType.indexOf("pdf") >= 0) {
		return "pdf";
	}
	if (mediaType.indexOf("xml") >= 0) {
		return "xml";
	}
	if (mediaType.indexOf("html") >= 0) {
		return "html";
	}
	if (mediaType.indexOf("json") >= 0) {
		return "json";
	}
	if (mediaType.indexOf("text") >= 0) {
		return "txt";
	}
	if (mediaType.indexOf("image") >= 0) {
		var ii = mediaType.lastIndexOf("/");
		if (ii >= 0) {
			return mediaType.slice(ii + 1);
		}
		return null;
	}
	if (mediaType.indexOf("msword") >= 0) {
		return "doc";
	}
	if (mediaType.indexOf("officedocument.word") >= 0) {
		return "docx";
	}
	if (mediaType.indexOf("ms-excel") >= 0) {
		return "xls";
	}
	if (mediaType.indexOf("officedocument.spreadsheet") >= 0) {
		return "xlsx";
	}
	if (mediaType.indexOf("ms-powerpoint") >= 0) {
		return "ppt";
	}
	if (mediaType.indexOf("officedocument.presentation") >= 0) {
		return "pptx";
	}
	return null;
}



function DocumentField() {}

exports.DocumentField = helpers.defineClass(DocumentField, Field, {
	loadBox: function() {
		this.isFieldTipDisabled = true;
		this.$authoringType = "documentField";
		if (this.isImage && this.$item.$isTopLabelAlignment == null) {
			this.$item.$isTopLabelAlignment = true;
		}
		Field.prototype.loadBox.call(this);
	},
	isPDFFile: function() {
		return this.$field.$type == "application/pdf";
	},
	isEmbeddable: function() {
		return this.isImage || this.isPDFFile();
	},

	initialize: function() {
		this.isDocumentField = true;
		this.isImage = this.$field.$type.indexOf("image") == 0;
		if (this.isImage) {
			// split image/star in two to avoid regex catastrophy
			this.acceptFilter = "image/" + "*";
			if (this.$item.$inplace) {
				this.$item.$imageWidth = this.$item.$imageWidth || "22px";
				this.$item.$imageHeight = this.$item.$imageHeight || "22px";
			} else {
				if (this.$item.$isAutoSize == undefined) {
					if (this.page.$facet == "$query") {
						//test pour dom autrememtn toujours mette true
						this.$item.$imageHeight = this.$item.$imageHeight || "50px";
						if (!(this.articleParent && this.articleParent.isGridRecord)) {
							if (this.$item.$imageWidth == "100%") {
								delete this.$item.$imageWidth;
							}
							this.$item.$imageWidth = this.$item.$imageWidth || "50px";
						}
					} else {
						this.$item.$isAutoSize = true;
					}
				}
				if (!this.$item.$isAutoSize) {
					this.$item.$imageWidth = this.$item.$imageWidth || "100%";
					this.$item.$imageHeight = this.$item.$imageHeight || "150px";
				}
			}
		}
		this.$dataUrl = this.getDataType().$url;
	},
	applyMetaData: function(metaData, options) {
		if (metaData) {
			if (metaData.$url) {
				this.$dataUrl = metaData.$url;
			}
			if (metaData.$contentType) {
				this.$contentType = metaData.$contentType;
			}
		}
		Field.prototype.applyMetaData.call(this, metaData, options);
	},
	applyDesignMetaData: function(metaData, designing) {
		Field.prototype.applyDesignMetaData.call(this, metaData, designing);
		if (this.isImage && designing) {
			if (metaData.$imageWidth !== undefined) {
				this.$item.$imageWidth = metaData.$imageWidth;
			}
			if (metaData.$imageHeight !== undefined) {
				this.$item.$imageHeight = metaData.$imageHeight;
				if (this.anchor) {
					this.anchor.style.height = this.$item.$imageHeight;
				}

			}
			if (metaData.$isAutoSize !== undefined) {
				this.applyAutoSize(metaData);
			}
			if (this._file) {
				this._setDocumentFile(this._file);
			}
		}
	},
	setState: function(state) {
		Field.prototype.setState.call(this, state);
		if (state.$isDisabled !== undefined) {
			_disableButtons(this, state.$isDisabled);
		}
		if (state.$isReadOnly !== undefined) {
			_showBtn(this, this.selectBtn, !state.$isReadOnly);
			_showBtn(this, this.removeBtn, !state.$isReadOnly);
		}
	},
	_upload: function(files) {
		var self = this;
		self.page.externalAdapter.onBeforeUploadFiles({
			field: self,
			files: files,
			doEvent: function() {
				for (var ii = 0; ii < files.length; ii++) {
					var file = files[ii];
					if (_validateFile(self, file) && ((self.acceptFilter && (file.type.match(self.acceptFilter) || self.acceptFilter.match(file.type))) || !self.acceptFilter)) {
						_onBeginUpload(self, file); //to improve for multi upload
						syra_controller.upload(file, self.getDataType().$url, self.$contentType || self.getDataType().$contentType, self.articleParent, function(data, response) {
							self._onUploadSuccess(file, data, response);
						}, function(progress, isLoaded) {
							if (!isLoaded) {
								_onProgressUpload(self, file, progress);
							} else {
								self._onEndUpload(file);
							}
						}, function(data, response) {
							self.articleParent.applyChange(data, response);
						});
					}
				}
			}
		});
	},
	_onEndUpload: function(file) {
		var self = this;
		self._legend.textContent = file.name + " " + syra_local.fdocUploaded;
		_disableButtons(self, false);
		_onProgressUpload(self, file, 100);
		setTimeout(function() {
			if (self._progress) {
				syra_site.dom.removeChild(self._progress);
				delete self._progress;
			}
			delete self.progressValue;
			if (self._legend) {
				self._legend.textContent = syra_local.fdocDropSelectFile;
			}
		}, 1500);
	},
	setLegend: function(legend) {
		this._legend.style.display = "";
		this._legend.textContent = legend;
	},
	hideSelectButton: function(isHidden) {
		_showBtn(this, this.selectBtn, !isHidden);
		this._legend.style.display = isHidden ? "none" : "";
		this._bindDropArea(false);
		!isHidden && this._bindDropArea(true); // Is Safer to unbind before... to prevent double call
	},
	render: function() {
		if (this.$isEditMode && !this.$item.$isFilterMode) {
			this._renderEditMode();
		}
	},
	_bindInputFileChange: function(bind) {
		var self = this;
		if (bind) {
			$(self.inputFile).bind("change", function(args) {
				self._upload(this.files);
			});
		} else {
			if (self.inputFile) {
				$(self.inputFile).unbind();
			}
		}
	},
	_bindDropArea: function(bind) {
		var $skin = this.$skin + (this.isImage ? "-img" : "-file");
		var self = this;
		if (bind) {
			self.$$dropArea.bind("drop dragenter dragover", function(event) {
				switch (event.type) {
					case "drop":
						self._upload(event.originalEvent.dataTransfer.files);
						break;
					case "dragenter":
						this.className += " s-file-drop-in";
						break;
					case "dragleave":
						this.className = $skin + "-drop-target";
						break;
				}
				event.stopPropagation();
				event.preventDefault();
				return false;
			});
		} else {
			if (self.$$dropArea) {
				self.$$dropArea.unbind();
			}
		}
	},
	onSelectClick: function(btn) {
		this.inputFile.click();
	},
	onRemoveClick: function(btn) {
		var self = this;
		self.page.externalAdapter.onRemoveUploadedFile({
			field: self,
			doEvent: function() {
				syra_controller.deleteFile(self.$dataUrl, self.articleParent, function(data, response) {
					self.setDataValue(null);
					self.selectBtn.focus();
					self.articleParent.applyChange(data, response);
				});
			}
		});
	},
	_renderEditMode: function() {
		var $skin = this.$skin + (this.isImage ? "-img" : "-file");
		var dropArea = document.createElement("div");
		dropArea.className = $skin + "-drop-target";
		this._dataValue.className += " s-drop-target-slot";
		this.$$dropArea = $(this._dataValue.appendChild(dropArea));
		this._bindDropArea(true);

		this.anchor = document.createElement("a");
		this.anchor.className = $skin + "-link";
		this.anchor.setAttribute("disabled", true);
		this.anchor.syraOnClick = "onInputFieldClick";
		this.anchor.setAttribute("data-s-field-extra", this.id);
		dropArea.insertBefore(this.anchor, dropArea.firstChild);

		this._right = document.createElement("div");
		this._right.className = this.$skin + "-doc-right";
		dropArea.appendChild(this._right);

		this._info = document.createElement("div");
		this._info.className = $skin + "-info";

		this._legend = document.createElement("div");
		this._legend.className = $skin + "-legend";
		this._legend.textContent = syra_local.fdocDropSelectFile;
		if (!this.$item.$isCellChild) {
			this._right.appendChild(this._info);
			this._info.appendChild(this._legend);
		}
		_appendActions(this, this.$skin);
		this.isImage && _setImageAnchorDefaultSize(this);
	},
	_renderDetailMode: function() {
		var $skin = this.$skin + (this.isImage ? "-img" : "-file");
		this.anchor = document.createElement("a");
		this.anchor.className = $skin + "-link";
		this.anchor.syraOnClick = "onInputFieldClick";
		if (!this.$item.$inplace) {
			this._dataValue.className = this._dataValue.className.replace(this._$cssField + "-value-read", "s-field-value-img");
		}
		this._dataValue.insertBefore(this.anchor, this._dataValue.firstChild);
		if (this.isImage) {
			if (this.$item.$inplace) {
				this.domItem.style.textAlign = "center";
			}
			this.anchor.style.height = this.$item.$imageHeight;
		}
	},
	_onUploadSuccess: function(file, data, response) {
		if (!data) {
			this._setDocumentFile(file);
		} else {
			this.page.applyChange(data, response);
		}
	},
	setDataValue: function(value) {
		this.articleParent.setMustBeResized();
		this.currentValue = value;
		if (!this.$isEditMode && this.currentValue) {
			this._renderDetailMode();
		}
		this._setDocumentFile(this.currentValue);
	},
	resizeField: function() {
		var imgW = this._imgLoader && this._imgLoader.width;
		if (this._imageArea && imgW) {
			var linW = this.anchor.clientWidth;
			var percent = (imgW >= linW) ? (linW / imgW) : 1;
			this._imageArea.style.height = (this._imgLoader.height * percent) + "px";
			this._imageArea.style.width = "100%";
		}
	},
	_setDocumentFile: function(file) {
		if (file && !_validateFile(this, file)) {
			return false;
		}
		if (this.isImage) {
			_setImageFile(this, file);
		} else {
			_setFile(this, file);
		}
	},
	getDisplayText: function() {
		return this.$fileUrl || "";
	},
	focus: function() {
		if (this.anchor && !this.$isHidden && !this.$isDisabled) {
			this.anchor.focus();
			this.scrollToField();
			return true;
		}
		this.scrollToField();
		return false;
	},
	dispose: function() {
		if (this.$isEditMode) {
			this.selectBtn = this.removeBtn = null;
			this._bindInputFileChange(false);
			this._bindDropArea(false);
		}
		if (this._imgLoader) {
			this._imgLoader.onload = this._imgLoader.onerror = this._imgLoader = null;
		}
		this._progress = this._info = this._imageWait = this._legend = this._file = this._imageArea = this.anchor = this.imageUrl = null;
		if (this.uploadFrame) {
			this.uploadFrame.removeEventListener("load", this.uploadFrameHandler, false);
		}
		this.onuploadFrameLoad = null;
		Field.prototype.dispose.call(this);
	}
});
