// syracuse-ui/lib/field/fieldArticle.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Article = require("syracuse-ui/lib/article/article").Article;

function FieldArticle() {}

exports.FieldArticle = helpers.defineClass(FieldArticle, Article, {
	dispose: function() {
		syra_article.dispose(this);
	},
	applyChange: function(newData) {
		syra_article.applyChange(this, newData);
	},
	ensureDataSet: function() {
		return this.dataset = syra_fields.setDatasetValue(this, syra_fields.getDatasetValue(this) || []);
	},
	ensureLayoutMode: function() {
		syra_fields.ensureLayoutMode(this);
	},
	getTitle: function() {
		return syra_fields.getTitle(this);
	},
	setTitle: function($title) {
		syra_fields.setTitle(this, $title);
		if (this.domTitle) {
			syra_dom.hide(this.domTitle, this.getTitle().length == 0);
		}
	},
	setState: function(state) {
		syra_fields.setState(this, state);
	},
	applyDesignMetaData: function(metaData, designing) {
		return syra_fields.applyDesignMetaData(this, metaData, designing);
	},
	applyMetaData: function(metaData, options) {
		syra_fields.applyMetaData(this, metaData, options);
	},
});
