// syracuse-ui/lib/field/field.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var locale = require('syracuse-core/lib/locale');
var fieldHelper = require('./fieldHelper');
var fieldDiagnoseUtils = require('syracuse-ui/lib/diagnoses/fieldDiagnosesUtils');

function Field(){
}

exports.Field = helpers.defineClass(Field, null, {
    applyDesignMetaData: function(metaData, onAuthoring){
        if (metaData.$isHidden !== undefined) {
            this.setState(metaData);
        }
        if (onAuthoring !== false) {
            if (metaData.$isTopLabelAlignment !== undefined) {
                var curItemCss = this.$skin;
                if (this.$item.$isTopLabelAlignment) {
                    curItemCss += "-top";
                }
                this.$item.$isTopLabelAlignment = metaData.$isTopLabelAlignment;
                var newItemCss = this.$skinField;
                if (this.$item.$isTopLabelAlignment) {
                    newItemCss += "-top";
                }
                this._domItem.className = this._domItem.className.replace(curItemCss, newItemCss);
                if (this.domTitle) {
                    this.domTitle.className = this.domTitle.className.replace(curItemCss + "-title", newItemCss + "-title");
                }
                this._core.className = this._core.className.replace(curItemCss + "-core", newItemCss + "-core");
            }
        }
        if (metaData.$isTitleHidden !== undefined) {
            this.$item.$isTitleHidden = metaData.$isTitleHidden;
            if (!metaData.$isTitleHidden && !this.domTitle) {
                this.appendTitle();
                this.$$item.prepend(this.domTitle);
                this.setTitle(this.getTitle());
            }
        }
        if (this.domTitle) {
            if (metaData.$isTitleHidden !== undefined) {
                this.hideTitle(metaData.$isTitleHidden);
            }
            if (metaData.$isRightTextLabelAlignment !== undefined) {
                this.domTitle.style.textAlign = (this.$item.$isRightTextLabelAlignment = metaData.$isRightTextLabelAlignment) ? "right" : "left";
            }
            
        }
    },
    
    onInputFocusin: function(input, event){
        if (this.articleParent.onInputFocusin) {
            this.articleParent.onInputFocusin(this);
        }
        this.hasFocus = true;
        document.site.onChildFieldFocus(this.boxParent, true);
        if (this.diagSeverities) {
            this.page.diagnosesPanel.onNavigationClick(0, this);
        }
    },
    onInputFocusout: function(input, event){
        if (this.articleParent.onInputFocusout) {
            this.articleParent.onInputFocusout(this);
        }
        this.hasFocus = false;
        this.onFieldMouseEvent({
            type: "mouseleave"
        });
        document.site.onChildFieldFocus(this.boxParent, false);
    },
    getLocalize: function(){
        if (!this._localize) {
            this._localize = locale.resources(module)()
        }
        return this._localize;
    },
    getArticle: function(){
        if (!this.articleParent) {
            this.articleParent = this.boxParent.getArticle();
        }
        return this.articleParent;
    },
    appendDetailLink: function(value, $detailsLink){
        if (!this.$isDetailLinkIgnore) {
            if (this.boxPickers) {
                $(this.boxPickers).detach();
            }
            document.site.emptyDom(this.fieldValue)
            this.page.loadNewItem(this.fieldValue, {
                $bind: "$detail",
                $title: $detailsLink.$title = value + "",
                $category: "link",
                $skin: this.$skin + "-link"
            }, this.boxParent).setMenu($detailsLink);
            if (this.boxPickers) {
                this.fieldValue.appendChild(this.boxPickers);
            }
        }
    },
    hideTitle: function($isTitleHidden){
        this.domTitle.style.display = $isTitleHidden ? "none" : "";
    },
    _createValueContainer: function(){
        this.domValueSlot = document.createElement("div");
        this.domValueSlot.className = this._$cssField + "-value";
        if (this.$cssFieldType) {
            this.domValueSlot.className += " " + this.$cssFieldType;
        }
        this.$$dataValue = $(this._dataValue = this.domValueSlot);
        if (this.$item.$inplace) {
            this._core = this.domValueSlot;
            this._domItem.appendChild(this.domValueSlot);
        }
        else {
            this.domValueSlot = document.createElement("div");
            this.domValueSlot.className = this._$cssField + "-value-slot";
            this.domValueSlot.appendChild(this._dataValue);
            
            this._core = document.createElement("div");
            this._core.className = this._$cssTopField + "-core";
            this._core.appendChild(this.domValueSlot);
            this._domItem.appendChild(this._core)
        }
        if (this.$isEditMode) {
            this.fieldValue = this._dataValue;
        }
        else {
            this._dataValue.appendChild(this.fieldValue = document.createElement("div"));
            this.fieldValue.className = this._$cssField + "-value-read";
        }
    },
    _initCss: function(){
        this._$cssTopField = this._$cssField = (this.$item.$css) ? (this.$item.$css + " " + this.$skinField) : this.$skinField;
        if (this.$item.$isTopLabelAlignment) {
            this._$cssTopField += "-top";
        }
    },
    ensureEditMode: function(){
        this.$isEditMode = this.$item.$isEditMode;
        if (this.$isEditMode === undefined) {
            this.$isEditMode = this.articleParent.$isEditMode;
        }
    },
    loadBox: function($$item){
        this.boxPickersPadding = 0;
        this.articleParent = this.boxParent.getArticle();
        this.ensureEditMode();
        this.$skinField = this.$skin = this.$item.$skin || (this.$item.$inplace ? "s-inplace" : "s-field");
        if (this.$item.$inplace) {
            this.arrayLevel = "cell";
        }
        if (this.$isEditMode) {
            this.$skinInput = this.$skin + "-input";
            if (this.$item.$isTopLabelAlignment == null) {
                this.$item.$isTopLabelAlignment = true;
            }
        }
        this._initCss();
        
        if (this.$isEditMode) {
            this.inputId = this.id + "-input";
        }
        this.$$item = $$item || $(this.layoutSlot);
        if (!this.$item.$inplace) {
            this.$$item = $$item || $(document.createElement("div")).appendTo(this.$$item);
        }
        (this._domItem = this.$$item[0]).className += " " + this._$cssTopField;
        
        this.$rootLinks = this.$field;
        if (this.initialize) {
            this.initialize();
        }
        
        this.appendTitle();
        
        this._createValueContainer();
        this.render();
        
        this._domItem.setAttribute("data-s-field", this.id);
        this._domItem.setAttribute("data-s-field-item", this.$item.$bind || "");
        if (this._input) {
            if (this.$item.$isBorderVisible) {
                this._input.className += " s-input-border";
            }
            if (this.$item.$isCellChild) {
                this._input.style.width = "100%";
                this._dataValue.style.width = "100%";
            }
            this._input.setAttribute("data-s-field", this.id);
            this._input.setAttribute("id", this.inputId);
        }
        if (!this.$item.$inplace && !this.$isEditMode) {
            this.domValueSlot.style.display = 'inline-block';
        }
        
        this.setTitle(this.$item.$title || this.$field.$title);
        
        this.setDescription(this.$item.$description || this.$field.$description);
        this.setHelp(this.$item.$help || this.$field.$help);
        this.applyDesignMetaData(this.$item, false);
        this.setFieldWidth();
        
        this.setState(this.$field);
        if (this.$field.$fieldStyle || this.$field.$valueStyle || this.$field.$titleStyle) {
            if (this.page.externalAdapter.applyDesignStyle(this, this.$field, this.$item)) {
                this.applyFieldStyle(this.$field);
            }
        }
        this.articleParent.bind(this, this.$item.$bind);
        this.setMenus(this.$field);
        this.applyAutoSize(this.$item);
        this._ensureMenuPickerVisible();
        if (document.site.developpementMode) {
            fieldHelper.appendQualityAtt(this, this._dataValue);
            if (this._input) {
                fieldHelper.appendQualityAtt(this, this._input);
            }
        }
        this.ensureLayoutMode();
    },
    setFieldWidth: function(){
        if (this.$isEditMode && !this.$item.$inplace && this.domValueSlot && !this.$item.$isAutoSizeDisabled) {
            var widthValue = document.site.getFieldWidth(this.$field, true);
            this.domValueSlot.style.maxWidth = widthValue + "px";
            if (this.layoutParent && widthValue > this.layoutParent.fieldMaxWidth) {
                this.layoutParent.fieldMaxWidth = widthValue;
            }
        }
    },
    applyFieldStyle: function(metaData){
        fieldHelper.applyFieldStyle(this, metaData);
    },
    applyAutoSize: function(state){
        if (state.$isAutoSize !== undefined) {
            this._bindAutoSize(state.$isAutoSize);
            this.$item.$isAutoSize = state.$isAutoSize;
        }
    },
    releaseMode: function(onDispose){
        this.domTitle = this.$$mandatoryFlag = this.fieldValue = this._description = this._helpFlag = null;
        if (onDispose) {
            if (this._menusBox) {
                document.controller.disposeObject(this._menusBox);
            }
            this.menuPicker = this._menusBox = null;
        }
    },
    _bindAutoSize: function(bind){
        document.site[bind ? "addResizeListener" : "removeResizeListener"](this);
    },
    render: function(){
        if (this.$isEditMode) {
            if (this.$item.$rows) {
                this._input = document.createElement("textarea");
                this.isTextArea = true;
                this._input.setAttribute("rows", this.$item.$rows);
            }
            else {
                if (this.createHtmlInput) {
                    this._input = this.createHtmlInput();
                }
                else {
                    this._input = document.createElement("input");
                    this._input.setAttribute("type", "text");
                }
                if (this.$item.$maxLength) {
                    this._input.setAttribute("maxlength", this.$item.$maxLength);
                }
            }
            this.$$input = $(this._input);
            
            this._input.className = this.$skinInput;
            if (this.$item.$css) {
                this._input.className += " " + this.$item.$css;
            }
            this.fieldValue.appendChild(this._input);
            this._appendLinksPicker();
        }
    },
    _appendSpecialPicker: function(pickerType){
        var picker = this.createPicker(pickerType);
        picker.setAttribute("data-s-article", this.articleParent.id);
        picker.setAttribute("data-s-field", this.id);
        this.appendToPickerBox(picker);
    },
    _appendLinksPicker: function(){
        if (this.$rootLinks.$links) {
            if (this.$rootLinks.$links.$suggest) {
                this._isAutoComplete = true;
            }
            if (this.$rootLinks.$links.$tunnel) {
                this._appendSpecialPicker("tunnel");
            }
            if (this.$rootLinks.$links.$lookup) {
                this._appendSpecialPicker("lookup");
            }
        }
    },
    _onAutoCompleteOpen: function(){
        return this._autoCompleteList = document.createElement("div");
    },
    _onAutoCompleteClose: function(){
        if (this._autoCompleteList) {
            document.site.emptyDom(this._autoCompleteList);
            this._autoCompleteList = null;
        }
    },
    toggleAutoCompleteList: function(show){
        var self = this;
        
        if (self.getDataValue() !== "") {
            self.$$item.attr("id", self.id);
            self.focus();
            
            self.searchValue = self.getDataValue();
            
            if (self._autoCompletePopup) {
                self._onAutoCompleteOpen();
            }
            else {
                self._autoCompletePopup = self.boxParent.openDialog({
                    content: self,
                    $dialogMode: "popup",
                    $$dialog: $(self._onAutoCompleteOpen()),
                    position: {
                        my: "right top",
                        at: "right bottom",
                        of: self.$$input
                    },
                    onClose: function(){
                        self.searchValue = null;
                        self._autoCompletePopup = null;
                        self._onAutoCompleteClose();
                    }
                });
            }
        }
        else {
            if (self._autoCompletePopup) {
                self._autoCompletePopup.close();
            }
        }
    },
    onInputKeydown: function(input, event){
        var self = this;
        if (self._isAutoComplete) {
            // don't run search if 'Enter' button
            if (event.keyCode !== 13) {
                //stop any previous ongoing search request
                clearTimeout(self._searching);
                self.runAutoCompleteSearch();
            }
            else {
                // remove autocomplete list if opened
                if (self._autoCompletePopup) {
                    self._autoCompletePopup.close();
                }
            }
        }
    },
    runAutoCompleteSearch: function(event){
        var self = this;
        self._searching = setTimeout(function(){
            // search doesn't take whitespace into account
            if (self.getDataValue().replace(/\s+/g, '') !== "") {
            
                // if some research has already been done, proceed with the request if the value has changed
                if (self.searchValue) {
                
                    if (self.searchValue !== self.getDataValue()) {
                        self.toggleAutoCompleteList(true);
                    }
                    else {
                        // if empty field, close incremental search popup
                        if (self.searchValue == self.getDataValue() == "") {
                            if (self._autoCompletePopup) {
                                self._autoCompletePopup.close();
                            }
                        }
                    }
                }
                else {
                    // first research (self.searchValue not set yet
                    self.toggleAutoCompleteList(true);
                }
            }
            else {
                if (self._autoCompletePopup) {
                    self._autoCompletePopup.close();
                }
            }
            
        }, 300);
    },
    removeFromPickerBox: function(picker){
        if (picker) {
            (this._input ? this._input : this.fieldValue).style.paddingRight = (this.boxPickersPadding -= 18) + "px";
            document.site.removeDomChild(picker);
            if (this.boxPickersPadding == 0 && this.boxPickers) {
                document.site.removeDomChild(this.boxPickers);
                delete this.boxPickers;
            }
        }
    },
    appendToPickerBox: function(picker){
        if (!this.boxPickers) {
            this.boxPickers = document.createElement("div");
            if (this._input) {
                this.fieldValue.appendChild(this.boxPickers).className = this.$skin + "-pickers";
            }
            else {
                this._dataValue.appendChild(this.boxPickers).className = this.$skin + "-pickers";
            }
            if (this.boxPickersPadding) {
                this.boxPickers.style.paddingRight = this.boxPickersPadding + "px";
            }
            if (this.$isDisabled || this.$isReadOnly) {
                this.boxPickers.style.visibility = "hidden";
            }
        }
        this.boxPickers.appendChild(picker);
        this._updateBoxPickerPadding(true);
    },
    _updateBoxPickerPadding: function(add){
        if (add) {
            this.boxPickersPadding += 18;
        }
        else {
            this.boxPickersPadding -= 18;
        }
        (this._input ? this._input : this._dataValue).style.paddingRight = this.boxPickersPadding + "px";
    },
    createPicker: function(pickerType){
        var btn = document.createElement("a");
        var css = this.$skin + "-" + pickerType;
        if (this.$item.$css) {
            css = this.$item.$css + " " + css;
        }
        btn.className = css + "-picker";
        btn.setAttribute("data-s-picker", pickerType);
        return btn;
    },
    doClickPicker: function(){
    
    },
    onFieldMouseEvent: function(event){
        if (this.menuPicker) {
            document.site.toggleClass(this.menuPicker._item, "s-mouseenter", (event.type == "mouseenter"));
        }
        if (this.boxPickers) {
            document.site.toggleClass(this.boxPickers, "s-mouseenter", (event.type == "mouseenter"));
        }
    },
    onFieldInputEvent: function(event){
        var self = this;
        self.page.externalAdapter.onFieldEvent({
            field: self,
            event: event,
            doEvent: function(){
                //    onInputFocusin,onInputFocusout,onInputKeyup,onInputKeydown,onInputKeypress: function(input, event){
                var fnct = "onInput" + event.type.charAt(0).toUpperCase() + event.type.slice(1);
                if (self[fnct]) {
                    self[fnct](event.target, event);
                }
            }
        });
    },
    onClickPicker: function(btn, onSelectRecord){
        var self = this;
        document.site.onBeforClick();
        if (!self.authoringNode && !self.$isDisabled) {
            var pickerType = btn.getAttribute("data-s-picker");
            self.page.externalAdapter.onFieldClickPicker({
                field: self,
                pickerType: pickerType,
                doEvent: function(options){
                    switch (btn.getAttribute("data-s-picker")) {
                        case "lookup":
                            self.focus();
                            var $lookup = self.$lookup || self.$rootLinks.$links.$lookup;
                            var $result = $lookup.$result;
                            self.boxParent.openDialog({
                                article: self.articleParent,
                                $url: $lookup.$url,
                                onSelectRecord: onSelectRecord ||
                                function(selectedRecords){
                                    var record = selectedRecords[Object.keys(selectedRecords)[0]];//single selection
                                    var dataValue = $result ? record.dataset[$result] : record.dataset;
                                    self.onSelectRecordAction(dataValue);
                                }
                            });
                            break;
                        case "tunnel":
                            break;
                        case "menus":
                            break;
                        default:
                            self.doClickPicker(btn, options);
                            break;
                    }
                }
            });
        }
    },
    onSelectRecordAction: function(dataValue){
        var self = this;
        if (!self.$isDisabled && !self.$isReadOnly) {
            self.setDataValue(dataValue);
            self.focus();
            self.notifyFieldChange(self.currentValue);
        }
    },
    getDefaultTitle: function(){
        return this.getLocalize().f_defaultTitle;
    },
    getTitle: function(){
        if (this.titleText == null && this.$titleValue) {
            this.titleText = this.$titleValue || "";
            if (this.titleText.length > 0 && this.titleText[1] == "@") {
                this.titleText = this.boxParent._renderExpression(this.titleText);
            }
        }
        return this.titleText || "";
    },
    setTitle: function($title){
        this.$titleValue = $title;
        delete this.titleText;
        if (this.domTitle) {
            this.domTitle.textContent = this.getTitle();
        }
        else {
            if (this._input && this.$item.$isPlaceHolder) {
                this._input.setAttribute("placeholder", this.getTitle());
            }
        }
    },
    appendTitle: function(){
        if (!(this.$item.$inplace || this.$item.$isTitleHidden)) {
            this.domTitle = document.createElement("label");
            this.domTitle.className = this._$cssTopField + "-title";
            if (this.inputId) {
                this.domTitle.setAttribute("for", this.inputId);
            }
            this._domItem.appendChild(this.domTitle);
        }
    },
    
    setHelp: function($help){
        return;
        if ($help != undefined) {
            this.$help = $help;
            if (this.domTitle && this.getTitle().length > 0) {
                if (this.$help) {
                    if (!this._helpFlag) {
                        this._helpFlag = document.createElement("a");
                        this._helpFlag.className = "s-field-help-flag";
                    }
                }
                else {
                    if (this._helpFlag) {
                        document.site.removeDomChild(this._helpFlag);
                        delete this._helpFlag;
                    }
                }
            }
        }
    },
    setDescription: function($description){
        if ($description !== undefined) {
            var text = this.$description = $description || "";
            if (text.length > 0) {
                if (text[1] == "@") {
                    text = this.boxParent._renderExpression(text);
                }
                if (!this.$item.$inplace) {
                    if (!this._description) {
                        this._description = document.createElement("div");
                        this._description.className = this.$skinField + "-desc";
                        $(this.domValueSlot).after(this._description);
                    }
                    this._description.textContent = text;
                    this._description.style.display = "";
                }
            }
            else {
                if (this._description) {
                    this._description.style.display = "none";
                }
            }
        }
    },
    setState: function(state){
        if (!this.$item.$isFilterMode) {
            if (state.$isMandatory !== undefined) {
                this.$isMandatory = state.$isMandatory;
                if (this.$isEditMode) {
                    if (this.domTitle && this.getTitle().length > 0) {
                        document.site.toggleClass(this.domTitle, "s-mandatory", state.$isMandatory);
                        if (state.$isMandatory) {
                            if (!this._mandatoryFlag) {
                                this._mandatoryFlag = document.createElement("div");
                                this._mandatoryFlag.className = "s-field-mandatory-flag";
                                this.domTitle.appendChild(this._mandatoryFlag);
                                if (document.site.developpementMode) {
                                    this._dataValue.setAttribute("data-s-q-mandatory", true);
                                    if (this._input) {
                                        this._input.setAttribute("data-s-q-mandatory", true);
                                    }
                                }
                            }
                        }
                        else {
                            if (this._mandatoryFlag) {
                                document.site.removeDomChild(this._mandatoryFlag);
                                delete this._mandatoryFlag;
                                if (document.site.developpementMode) {
                                    this._dataValue.removeAttribute("data-s-q-mandatory");
                                    if (this._input) {
                                        this._input.removeAttribute("data-s-q-mandatory");
                                    }
                                }
                            }
                        }
                    }
                    document.site.toggleClass(this._dataValue, "s-mandatory", state.$isMandatory);
                }
            }
            if (state.$isDisabled !== undefined) {
                this.$isDisabled = state.$isDisabled;
                this.$$item.attr("disabled", state.$isDisabled);
                if (this._input) {
                    document.site.disableItem(this._input, state.$isDisabled);
                }
                if (this.domTitle) {
                    document.site.disableItem(this.domTitle, state.$isDisabled);
                }
                if (this._description) {
                    document.site.disableItem(this._description, state.$isDisabled);
                }
                document.site.disableItem(this._dataValue, state.$isDisabled);
                if (this.boxPickers) {
                    this.boxPickers.style.visibility = (this.$isDisabled || this.$isReadOnly) ? "hidden" : "";
                }
                if (this.menuPicker) {
                    this.menuPicker._item.style.visibility = this.$isDisabled ? "hidden" : "";
                }
            }
            if (state.$isReadOnly !== undefined) {
                this.$isReadOnly = state.$isReadOnly;
                if (this._onSetReadOnly) {
                    this._onSetReadOnly(state.$isReadOnly);
                }
                if (this.boxPickers) {
                    this.boxPickers.style.visibility = (this.$isDisabled || this.$isReadOnly) ? "hidden" : "";
                }
            }
        }
        if (state.$isHidden !== undefined) {
            this.$isHidden = this.$item.$isHidden = state.$isHidden; //at root for fusion
            if (this.$item.$inplace) {
                this.$$dataValue[0].style.display = state.$isHidden ? "none" : "";
            }
            else {
                this.layoutSlot.style.display = this.$$item[0].style.display = state.$isHidden ? "none" : "";
                
            }
        }
    },
    _onSetReadOnly: function($isReadOnly){
        if (this._input) {
            if (this.$isReadOnly) {
                this._input.setAttribute("readonly", "readonly");
            }
            else {
                this._input.removeAttribute("readonly");
            }
            document.site.toggleClass(this._input, "s-readonly", this.$isReadOnly);
        }
    },
    setDataBind: function(value, record, metaData){
        if (this.page.externalAdapter.setDataBind(this, value, record, metaData)) {
            if (value !== undefined) {
                this.dirty(false);
                this.setDataValue(value, record);
            }
            this.applyMetaData(metaData);
        }
    },
    setDataValue: function(value, record){
        this.currentValue = value;
        if (this.$isEditMode) {
            this._input.value = this.currentValue || "";
            // quick hack for lookup in regular field
            if (this.currentValue && typeof(this.currentValue) == 'object' && this.$item && this.$item.$bind) {
                this._input.value = this.articleParent.parseExpression("{" + this.$item.$bind + "}", value);
                this.currentValue.$value = this._input.value;
            }
        }
        else {
            document.site.emptyDom(this.fieldValue);
            if (value && this.$rootLinks.$links && this.$rootLinks.$links.$details && !this.$item.$isMenusDisabled) {
                this.appendDetailLink(value, this.$rootLinks.$links.$details);
            }
            else {
                this.fieldValue.textContent = value;
            }
        }
    },
    dirty: function(set){
        if (set !== undefined) {
            this._isDirty = set;
        }
        else 
            return this.isTextArea ? this._isDirty : true;
    },
    //old getinputValue
    getDataValue: function(){
        return this.getInputValue ? this.getInputValue() : this._input.value;
    },
    
    /// -------------
    /// ## getCaretPosition()
    /// 
    /// Gets the current 0 based caret position in the input.
    /// 
    getCaretPosition: function(){
        return this.$item && this.$item.$rows && this.$$input && (this.$$input.prop("selectionDirection") === "forward" ? this.$$input.prop("selectionEnd") : this.$$input.prop("selectionStart"));
    },
    
    
    applyMetaData: function(metaData, options){
        if (metaData) {
            if (metaData.$isEditMode !== undefined) {
                this.toggleEditMode(metaData.$isEditMode);
            }
            if (metaData.$title !== undefined) {
                this.setTitle(metaData.$title);
            }
            if (metaData.$description !== undefined) {
                this.setDescription(metaData.$description);
            }
            if (metaData.$help !== undefined) {
                this.setHelp(metaData.$help);
            }
            this.setState(metaData);
            if (metaData.$diagnoses !== undefined) {
                this.showDiagnoses(metaData.$diagnoses, options);
            }
            this.setMenus(metaData);
            if (metaData.$links) {
                if (this.$isEditMode) {
                    if (metaData.$links.$lookup) {
                        if (!this.$lookup) {
                            this._applyLookUpLinkChange(metaData.$links.$lookup);
                        }
                    }
                }
                else {
                    if (metaData.$links.$details) {
                        if (this.currentValue && !this.$item.$isMenusDisabled) {
                            this.appendDetailLink(this.currentValue, metaData.$links.$details);
                        }
                    }
                }
            }
            
            if ((metaData.$type !== undefined) && (this.$field.$type != metaData.$type)) {
                fieldHelper.changeType(metaData.$type, this);
            }
            //$valueStyle can be "" (clear style)  in this case  metaData.$valueStyle is false
            if (metaData.$fieldStyle || metaData.$valueStyle != undefined || metaData.$titleStyle) {
                this.applyFieldStyle(metaData);
            }
        }
        if (this._ensureMenuPickerVisible) {
            this._ensureMenuPickerVisible();
        }
    },
    _applyLookUpLinkChange: function($newLookup){
        var self = this;
        if (!self.$lookup) {
            if (self.$rootLinks.$links && self.$rootLinks.$links.$lookup) {
                self.$lookup = helpers.object.clone(self.$rootLinks.$links.$lookup, true);
            }
            else {
                self.$lookup = $newLookup;
                return;
            }
        }
        Object.keys($newLookup).forEach(function($property){
            self.$lookup[$property] = $newLookup[$property];
        });
    },
    focus: function(){
        if (this._input) {
            if (!(this.$isHidden || this.$isDisabled)) {
                this._input.focus();
                return true;
            }
        }
        return false;
    },
    isEmpty: function(){
        return !this.currentValue;
    },
    getDataType: function(){
        return this.$field;
    },
    toggleDiagnose: function(css, show){
        fieldDiagnoseUtils.toggleDiagnose(this, css, show);
    },
    drawDiagnose: function($diagnose, severityGroup){
        fieldDiagnoseUtils.drawDiagnose(this, $diagnose, severityGroup);
    },
    emptyDiagnoseSlot: function(){
        fieldDiagnoseUtils.emptyDiagnoseSlot(this);
    },
    ensureDiagnoseSlot: function(){
        fieldDiagnoseUtils.ensureDiagnoseSlot(this);
    },
    showErrors: function(errors, options){
        this.showDiagnoses((this.$displayedDiagnoses = errors.map(function(error){
            return {
                message: error,
                severity: "error"
            };
        })), options);
    },
    showDiagnoses: function(diagnoses, options){
        if (!this.$item.$isExpressionChild) {
            document.site.showDiagnoses({
                field: this,
                $diagnoses: (diagnoses && diagnoses.length > 0) ? diagnoses : null
            }, this.boxParent, options);
        }
    },
    validate: function(newValue, errors){
        if (this.$item.$isFilterMode) {
            return true;
        }
        var value = newValue;
        if (value === undefined) {
            value = this.currentValue;
        }
        if (value == null) {
            value = "";
        }
        value = value.toString();
        var dataType = this.getDataType();
        errors = errors || [];
        if (this.$isMandatory && value == '' && this.page.externalAdapter.notifyFldMandatoryErr()) {
            errors.push(this.getLocalize().f_valueIsMandatory);
        }
        if (dataType.$pattern && value) {
            var match = new RegExp(dataType.$pattern, "i").test(value);
            if (!match) { //(match && match.index == 0 && match[0].length == value.length)) {
                errors.push(this.getLocalize().f_invalidValue + " : " + value);
            }
        }
        if (this.validateType) {
            this.validateType(errors, value);
        }
        
        if (this.articleParent.onFieldValidate) {
            this.articleParent.onFieldValidate(this, value, errors);
        }
        if (errors.length > 0 || this.$displayedDiagnoses) {
            this.showErrors(errors);
            var saveLinks = this.articleParent.menuItems.$save;
            if (saveLinks && saveLinks.length > 0) {
                if (!saveLinks[0].$isDisabled) {
                    this.$isServerSaveEnabled = true;
                }
                else {
                    if (this.$isServerSaveEnabled) {
                        this.$isServerSaveEnabled = false;
                    }
                }
                if (this.$isServerSaveEnabled !== undefined) {
                    saveLinks[0].disable(this.$isServerSaveEnabled);
                }
            }
        }
        return errors.length == 0;
    },
    onInputValidate: function(event, notify){
        var newValue = this.getDataValue(event);
        if (notify || (newValue != this.currentValue)) {
            this.notifyFieldChange(newValue);
        }
    },
    onInputChange: function(input, event){
        this.dirty(true);
        if (this.$item.$validationTrigger !== 'keyup') {
            this.onInputValidate(event);
        }
    },
    onInputKeyup: function(input, event){
        if (this.$item.$validationTrigger == 'keyup' && event.keyCode == 13) {
            this.onInputValidate(event, true);
        }
    },
    notifyFieldChange: function(newValue, validated){
        var self = this;
        self.currentValue = newValue;
        if (validated || self.validate()) {
            self.articleParent.dataset[self.$item.$bind] = self.currentValue;
            self.page.externalAdapter.onFieldNotifyChange({
                field: self,
                doEvent: function(){
                    if (self.articleParent.onNotifyDataChange ? self.articleParent.onNotifyDataChange(self, self.currentValue) : true) {
                        self.page.notifyDataChange(self, self.currentValue);
                    }
                }
            });
        }
    },
    toggleEditMode: function($isEditMode){
        fieldHelper.toggleEditMode(this, $isEditMode);
    },
    setMenus: function(metaData){
        if (!this.$item.$isFilterMode && !this.$item.$isExpressionChild) {
            var $menus = metaData.$item || metaData;
            if ($menus.$links === null) {
                if (this._menusBox) {
                    this._menusBox.unload();
                    this.menuPicker = this._menusBox = null;
                    this._ensureMenuPickerVisible(true);
                }
            }
            else {
                if ($menus.$links !== undefined || $menus.$actions !== undefined) {
                    this._ensureMenusBox().ensurePrivateMenus($menus);
                }
            }
        }
    },
    filterMenu: function($bind){
        switch ($bind) {
            case "$details":
                if (!this.$isDetailLinkIgnore) {
                    return false;
                }
                break;
            case "$tunnel":
                return this.$item.$inplace && !this.$item.$isFilterMode;
            case "$suggest":
                this._isAutoComplete = true;
                return false;
            case "$lookup":
            case "$select":
                return false;
            case "$lazyload":
                this._menusBox.isLazyLoad = true;
                return false;
        }
        return true;
    },
    /*if (!this.$item.$inplace) {
     if (this.$rootLinks.$links.$tunnel) {
     this._appendSpecialPicker("tunnel");
     }
     }*/
    _ensureMenusBox: function(){
        if (!this._menusBox) {
            this._menusBox = this.page.loadNewItem(this._input ? this.fieldValue : this._dataValue, {
                $isOwner: true,
                $id: this.id + "-menu-picker",
                $skin: "s-field-context-menus",
                $itemSkin: "s-field-menus-link",
                $category: "menus",
                $isBoxCollapsable: true,
                $isPopupContent: true,
                $title: "-"
            }, this.articleParent);
            this._menusBox.contextField = this;
            this.menuPicker = this.articleParent.idMap[this.id + "-menu-picker"];
        }
        return this._menusBox;
    },
    _isShowMenuPicker: function(){
        var show = (this.$isEditMode || (this.currentValue != null));
        return (show && (this.menuPicker.isLazyLoad || this.menuPicker.$item.$layout.$items.length > 0));
    },
    _ensureMenuPickerVisible: function(isRemoved){
        var show;
        if (this.menuPicker) {
            show = this._isShowMenuPicker();
            if (show && !this.menuPicker.$isPickerVisible) {
                this.menuPicker.$isPickerVisible = true;
                this.menuPicker._item.style.display = "";
                if (this.boxPickers) {
                    this.boxPickers.style.paddingRight = "18px";
                }
                if (this.$isDisabled) {
                    this.menuPicker._item.style.visibility = "hidden";
                }
                this._updateBoxPickerPadding(true);
            }
        }
        if (isRemoved || show === false) {
            if (this.boxPickers) {
                this.boxPickers.style.paddingRight = "";
            }
            if (this.menuPicker) {
                this.menuPicker._item.style.display = "none";
                if (this.menuPicker.$isPickerVisible) {
                    this.menuPicker.$isPickerVisible = false;
                    this._updateBoxPickerPadding(false);
                }
            }
            else {
                this._updateBoxPickerPadding(false);
            }
        }
    },
    ensureLayoutMode: function(){
        if (this.layoutParent) {
            this.layoutParent.ensureLayoutSlot(this);
        }
        if (this.page.authoringPage) {
            this.page.authoringPage.toggleItemAuthoring(this, true);
        }
    },
    dispose: function(){
        this.releaseMode(true);
        if (this.$item) {
            if (this.$item.$isAutoSize) {
                this._bindAutoSize(false);
            }
        }
        this.diagSeverities = this.layoutParent = this.authoringNode = this.boxParent = this.articleParent = null;
        this.page = this.boxPickers = this.domValueSlot = null;
        this.$field = this.$item = this.boxParent = null;
        this.layoutSlot = this.$$item = this._domItem = this.$$dataValue = this._dataValue = null;
        this._core = this.$$input = this._input = this.currentValue = this.menuPicker = this._menusBox = null;
        this._autoCompleteList = null;
    }
    
});
