// syracuse-ui/lib/field/fieldHelper.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

exports.appendQualityAtt = function(field, dom){
    dom.setAttribute("data-s-q-bind", field.$item.$bind);
    dom.setAttribute("data-s-q-title", field.getTitle());
    if (field.$field.$X3Name) {
        dom.setAttribute("data-s-q-x3Name", field.$field.$X3Name);
    }
};
exports.changeType = function($newType, oldField){
    var article = oldField.articleParent;
    var $oldField = article.$prototype.$properties[oldField.$item.$bind];
    if (typeof($newType) == "string") {
        $oldField.$type = $newType;
    }
    else {
        $oldField = $newType;
    }
    var newField = oldField.page.createNewItem(oldField.$item, oldField.boxParent);
    newField = oldField.layoutParent.createChildItem(oldField.$item);
    
    oldField.layoutParent.createChildItem({
        newItem: newField,
        load: oldField.$$item != null,
        targetItem: oldField,
        action: "insertAfter"
    });
    var boundFields = article.boundFields[newField.$item.$bind];
    boundFields.splice(boundFields.indexOf(newField), 1);
    boundFields.splice(boundFields.indexOf(oldField), 1, newField);
    article.removeItem(oldField, true);
};

exports.toggleEditMode = function(field, $isEditMode){
    if ($isEditMode === undefined) {
        $isEditMode = !field.$isEditMode;
    }
    if (field.$isEditMode != $isEditMode) {
        field.$isEditMode = field.$item.$isEditMode = $isEditMode;
        var $$newItem;
        if (field.$item.$inplace) {
            document.site.emptyDom(field.layoutSlot);
        }
        else {
            field.$$item.replaceWith($$newItem = $(document.createElement("div")));
        }
        field.releaseMode();
        field.loadBox($$newItem);
        
        field.setTitle(field.getTitle());
        field.setDescription(field.$description);
        field.setState(field);
        
        //field._restoreButtonValue();
    }
};
function _updateFusionStyle(field, saveKey, target, newStyle){
    var previous = field[saveKey];
    if (previous) {
        //clean
        var domStyle = target.getAttribute("style");
        var domStyleParts = domStyle.split(";");
        for (var mm = 0, kk = domStyleParts.length; mm < kk; mm++) {
            domStyleParts[mm] = domStyleParts[mm].trim();
        }
        var previousParts = previous.split(";");
        for (var ii = 0, jj = previousParts.length; ii < jj; ii++) {
            if (previousParts[ii]) {
                var parts = previousParts[ii].split(":");
                if (parts.length > 0) {
                    parts[0] = parts[0].trim();
                    for (var mm = 0, kk = domStyleParts.length; mm < kk; mm++) {
                        if (domStyleParts[mm].indexOf(parts[0]) == 0) {
                            domStyleParts.splice(mm, 1);
                            break;
                        }
                    }
                }
            }
        }
        target.setAttribute("style", domStyleParts.join(";"));
    }
    if (newStyle) {
        field[saveKey] = newStyle;
        target.setAttribute("style", (target.getAttribute("style") || "") + ";" + newStyle);
    }
    else {
        delete field[saveKey];
    }
}

function _updateStyle(field, saveKey, target, newStyle){
    if (field[saveKey]) {
        document.site.toggleClass(target, field[saveKey], false);
    }
    document.site.toggleClass(target, (field[saveKey] = (newStyle || "").replace("cst_sty_", "s-cst-sty-")), true);
}

exports.applyFieldStyle = function(field, metaData){
    var target;
    var newStyle = metaData.$fieldStyle || metaData.$valueStyle;
    if (newStyle !== undefined) {
        target = field._input ? field._input : field._dataValue;
        if (field.page.$isFusionPage) {
            _updateFusionStyle(field, "$valueStyle", target, newStyle);
        }
        else {
            _updateStyle(field, "$valueStyle", target, newStyle);
        }
    }
    newStyle = metaData.$fieldStyle || metaData.$titleStyle;
    if (field.domTitle && newStyle !== undefined) {
        if (field.page.$isFusionPage) {
            _updateFusionStyle(field, "$titleStyle", field.domTitle, newStyle);
        }
        else {
            _updateStyle(field, "$titleStyle", field.domTitle, newStyle);
        }
    }
};
