// syracuse-ui/lib/field/fieldHelper.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

exports.appendQualityAtt = function(field, dom){
    dom.setAttribute("data-s-q-bind", field.$item.$bind);
    dom.setAttribute("data-s-q-title", field.getTitle());
    if (field.$field.$X3Name) {
        dom.setAttribute("data-s-q-x3Name", field.$field.$X3Name);
    }
};
exports.changeType = function($newType, oldField){
    var article = oldField.getArticleParent ? oldField.getArticleParent() : oldField.getArticle();
    var $oldField = article.$prototype.$properties[oldField.$item.$bind];
    if (typeof($newType) == "string") {
        $oldField.$type = $newType;
    }
    else {
        $oldField = $newType;
    }
    var newField = oldField.page.createNewItem(oldField.$item, oldField.boxParent);
    oldField.layoutParent.appendNewItem({
        newItem: newField,
        load: oldField.$$item != null,
        targetItem: oldField,
        action: "insertAfter"
    });
    var boundFields = article.boundFields[newField.$item.$bind];
    boundFields.splice(boundFields.indexOf(newField), 1);
    boundFields.splice(boundFields.indexOf(oldField), 1, newField);
    article.removeItem(oldField, true);
};


exports.toggleEditMode = function(field, $isEditMode){
    if ($isEditMode === undefined) {
        $isEditMode = !field.$isEditMode;
    }
    if (field.$isEditMode != $isEditMode) {
        field.$isEditMode = field.$item.$isEditMode = $isEditMode;
        var $$newItem;
        if (field.$item.$inplace) {
            field.$$container.empty();
        }
        else {
            field.$$item.replaceWith($$newItem = $(document.createElement("div")));
        }
        if (field._$$buttons) {
            field._$$buttons.detach();
        }
        field.releaseMode();
        field.loadBox($$newItem);
        
        field.setTitle(field.getTitle());
        field.setDescription(field.$description);
        field.setState(field);
        
        //field._restoreButtonValue();
    }
};

exports.applyFieldStyle = function(field, metaData){
    var target;
    if (metaData.$fieldStyle !== undefined || metaData.$valueStyle !== undefined) {
        target = field._input ? field._input : field._dataValue;
        if (field.page.$isFusionPage) {
            if (field.$valueStyle) {
                var curStyle = target.getAttribute("style") || "";
                target.setAttribute("style", curStyle.replace(field.$valueStyle, ""));
            }
            if (field.$valueStyle = metaData.$fieldStyle || metaData.$valueStyle) {
                var curStyle = target.getAttribute("style") || "";
                target.setAttribute("style", curStyle + ";" + field.$valueStyle);
            }
        }
        else {
            if (field.$valueStyle) {
                document.site.toggleClass(target, field.$valueStyle, false);
            }
            field.$valueStyle = (metaData.$fieldStyle || metaData.$valueStyle || "").replace("cst_sty_", "s-cst-sty-");
            document.site.toggleClass(target, field.$valueStyle, true);
        }
    }
    if (field.domTitle && (metaData.$fieldStyle !== undefined || metaData.$titleStyle !== undefined)) {
        if (field.page.$isFusionPage) {
            target = field._input ? field._input : field._dataValue;
            if (field.$titleStyle) {
                var curStyle = field.domTitle.getAttribute("style") || "";
                field.domTitle.setAttribute("style", curStyle.replace(field.$titleStyle, ""));
            }
            if (field.$titleStyle = metaData.$fieldStyle || metaData.$titleStyle || "") {
                var curStyle = target.getAttribute("style") || "";
                field.domTitle.setAttribute("style", curStyle + ";" + field.$titleStyle);
            }
        }
        else {
            if (field.$titleStyle) {
                document.site.toggleClass(field.domTitle, field.$titleStyle, false);
            }
            field.$titleStyle = (metaData.$fieldStyle || metaData.$titleStyle || "").replace("cst_sty_", "s-cst-sty-");
            document.site.toggleClass(field.domTitle, field.$titleStyle, true);
        }
    }
};

/********************** Diagnose Management  ***********************************************/
exports.drawDiagnose = function(field, $diagnose, severityGroup){
    var link = document.createElement("a");
    
    var label = document.createElement("label");
    label.className = "s-field-diagnose-msg-" + severityGroup.severity;
    
    if (field.isDiagnoseInPlace) {
        if (!field._diagnoseFlag) {
            field._diagnoseFlag = document.createElement("a");
            field._diagnoseFlag.className = "s-field-diagnose-flag-" + severityGroup.severity;
            field.appendToPickerBox(field._diagnoseFlag);
        }
        link.className = "s-field-diagnose-link s-field-diagnose-link-" + severityGroup.severity;
        label.innerHTML = document.site.formatHTMLMessage($diagnose.message);
        
        $(label).prepend($(link).text(field.getTitle() + ":")).appendTo(field.domDiagnose);
    }
    else {
        if (!severityGroup.$$group.children().children("a[data-s-menu='" + field.id + "']")[0]) {
            link.className = "s-diag-view-msg-link";
            severityGroup.drawLabel($diagnose).prepend($(link).text(field.$field.$title + ":")).appendTo(severityGroup.$$group);
            drawBoxParentDiagnose(field, $diagnose);
        }
        label.innerHTML = document.site.formatHTMLMessage($diagnose.message);
        field.domDiagnose.appendChild(label);
        link.setAttribute("data-s-menu", field.id);
    }
    link.setAttribute("data-s-article", field.articleParent.id);
    link.setAttribute("data-s-field", field.id);
};





function TitleDiagnose($$titleDom){
    this.$$titleDom = $$titleDom;
    this.severities = {};
}

helpers.defineClass(TitleDiagnose, null, {
    addSeverityCount: function(severity){
        var self = this;
        // add severity count
        if (!self.severities[severity].count) {
            self.severities[severity].count = 1;
        }
        else {
            self.severities[severity].count++;
        }
    },
    removeSeverityCount: function(severity){
        var self = this;
        self.severities[severity].count--;
    },
    getPrioritySeverity: function(){
        var max = 0;
        var severity = "error";
        for (s in this.severities) {
            if (s.count > max) {
                max = s.count;
                severity = s;
            }
        }
        return severity;
    },
    setSeverity: function(){
    
    },
    appendDiagnoseIcon: function(severity, $isTab){
        var self = this;
        // add severity count
        self.addSeverityCount(severity);
        
        var diagClassNameRoot = $isTab ? "s-box-tab-" : "s-box-collapsible-";
        var diagClassName = diagClassNameRoot + severity;
        
        // icon already appended
        if (self.$hasIcon) {
            // severity priority change ?
            var prioritySeverity = self.getPrioritySeverity();
            if (self.prioritySeverity != prioritySeverity) {
                self.$$titleDom.toggleClass(diagClassNameRoot + self.prioritySeverity).toggleClass(diagClassNameRoot + prioritySeverity);
                self.prioritySeverity = prioritySeverity;
            }
        }
        // icon not yet appended.
        else {
            self.$$titleDom.addClass(diagClassNameRoot + severity);
            self.$hasIcon = true;
            self.prioritySeverity = severity;
        }
        
        // add severity count first
        // two possibilites 
        //"icon already" -> change icon or not ?
        //"icon not yet" -> add it
    
    },
    removeDiagnoseIcon: function(){
        // only if count = 0
    },
    dispose: function(){
        delete this.severities;
    }
});




function drawBoxParentDiagnose(field, $diagnose){
    var f = field;
    
    while (f.boxParent) {
        // collapsible
        //f.boxParent.$opened
        if (f.boxParent.$item.$isBoxCollapsable) {
            (f.boxParent._header.titleDiagnose = f.boxParent._header.titleDiagnose || new TitleDiagnose($(f.boxParent._header))).appendDiagnoseIcon($diagnose.severity);
        }
        
        else 
            // tab 
            //f.boxParent.$isTabLayout
            if (f.boxParent._tabTitle) {
                //$(f.boxParent._tabTitle).toggleClass("s-tab-header-error");
                (f.boxParent._tabTitle.titleDiagnose = f.boxParent._tabTitle.titleDiagnose || new TitleDiagnose($(f.boxParent._tabTitle))).appendDiagnoseIcon($diagnose.severity, true);
            }
        f = f.boxParent;
    }
};

function emptyBoxParentDiagnose(field){
    var f = field;
    
};

exports.emptyDiagnoseSlot = function(field){
    if (field.domDiagnose) {
        document.site.emptyDom(field.domDiagnose);
        field.domDiagnose.style.visibility = "hidden";
        if (field._diagnoseFlag) {
            field.removeFromPickerBox(field._diagnoseFlag);
            delete field._diagnoseFlag;
        }
    }
    //TODO call emptyBoxParentDiagnose();
    // in the function, decrease count. if 0, toggleClass
    emptyBoxParentDiagnose(field);
};
exports.ensureDiagnoseSlot = function(field){
    if (!field.domDiagnose) {
        field.domDiagnose = document.createElement("div");
        field.domDiagnose.className = "s-field-diagnose";
        if (field.$item.$inplace && field.articleParent.ensureFieldDiagnoseSlot) {
            field.isDiagnoseInPlace = true;
            field.articleParent.ensureFieldDiagnoseSlot(field);
        }
        else {
            field._core.appendChild(field.domDiagnose);
        }
    }
    else {
        document.site.emptyDom(field.domDiagnose);
    }
};

exports.showErrors = function(field, errors, options){
    field.applyMetaData({
        $diagnoses: null
    }, options);
    if (errors.length > 0) {
        field.applyMetaData({
            $diagnoses: (field.$displayedDiagnoses = errors.map(function(error){
                return {
                    message: error,
                    severity: "error"
                };
            }))
        }, options);
    }
};

exports.showDiagnoses = function(field, diagnoses, options){
    field.applyMetaData({
        $diagnoses: null
    }, options);
    if (diagnoses.length > 0) {
        field.applyMetaData({
            $diagnoses: diagnoses
        }, options);
    }
};

exports.toggleDiagnose = function(field, css, show){
    if (field._input) {
        document.site.toggleClass(field._input, css, show);
    }
    document.site.toggleClass(field._dataValue, css, show);
};
