// syracuse-ui/lib/field/numericField.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Field = require('syracuse-ui/lib/field/field').Field;
var _formatApi = require('syracuse-ui/lib/field/formatApi');

function _formatInput(field, val){
    var type = field.$field.$type;
    var numFormatObj = field.numFormatObj;
    var ds = numFormatObj.decimalSeparator;
    var gs = numFormatObj.groupSeparator;
    
    var chars_ok = "0-9-+";
    
    // for decimal values, removing invalid chars and setting '.' as decimal separator (javascript float)
    if (type == "application/x-decimal" || type == "application/x-real") {
    
        chars_ok += ds;
        
        if (gs != ".") {
            var regexp = new RegExp("[^" + chars_ok + "." + "]", "g");
            val = val.replace(regexp, '');
        }
        else {
            var regexp = new RegExp("[^" + chars_ok + "]", "g");
            val = val.replace(regexp, '');
        }
        
        if (ds != ".") {
            var regexp = new RegExp(ds, "g");
            val = val.replace(regexp, '.');
        }
    }
    // for integer values, removing invalid chars
    else {
        var regexp = new RegExp("[^" + chars_ok + "]", "g");
        val = val.replace(regexp, '');
    }
    
    val = type == "application/x-decimal" ? parseFloat(val) : parseInt(val, 10);
    
    // control
    if (isNaN(val)) {
        throw new Error("invalid number : " + val);
    }
    
    // formatting
    return field.formatApi.format(field.formatApi.parse(val), field.localFormat);
}


function NumericField(){
}

exports.NumericField = helpers.defineClass(NumericField, Field, {
    format: function(val, format){
        return this.formatApi.format(val, format);
    },
    validateType: function(errors, value){
    
    },
    dispose: function(){
        var formatEx;
        delete this.formatApi;
        delete this.numFormatObj;
        delete this.localFormat;
        delete this._formatInput;
        delete this._checkSep;
        if ((formatEx = this.formatEx)) {
            Object.keys(formatEx).forEach(function(key){
                delete formatEx[key];
            });
        }
        Field.prototype.dispose.call(this);
    },
    initialize: function(){
        this.formatApi = _formatApi.getApi(this.$field.$type);
        this.numFormatObj = _formatApi.getLocale().getNumberFormatObj(this.$field.$type);
        this.localFormat = this.$field.$format || this.numFormatObj.numFormat;
        this._formatInput = _formatInput;
        this._checkSep = true;
        //this.localFormat = _formatApi.getLocalFormat(this.$field.$type, this.$field.$format);
        if (this.formatEx) {
            // Overwrite format APIs  "formatApi", "localFormat", etc...
            helpers.object.extendEx(this, this.formatEx);
            this.formatApi.regionalOptions = this.numFormatObj;
        }
    },
    onInputFocusin: function($$input, event){
        var self = this;
        $$input.toggleClass('s-field-input-error', false);
        var gs = self.numFormatObj.groupSeparator;
        //var ds = self.numFormatObj.decimalSeparator;
        var gsRegExp = gs == " " ? new RegExp("\\s", "g") : new RegExp("[" + gs + "]", "g");
        //var dsRegExp = new RegExp("["+ds+"]","g");
        var editVal = $$input.val().split(gsRegExp).join('');
        this.$$input.val(editVal);
        Field.prototype.onInputFocusin.call(this, $$input, event);
    },
    onInputFocusout: function($$input, event){
        try {
            var self = this;
            if (this._checkSep && !self.valHasChanged) {
                var ds = self.numFormatObj.decimalSeparator;
                var rawNum = ds != '.' ? $$input.val().replace(new RegExp("[" + ds + "]"), '.') : $$input.val();
                var numObj = self.formatApi.parse(rawNum);
                // displaying formatted num
                self.$$input.val(self.formatApi.format(numObj, self.localFormat))
            }
            else {
                self.valHasChanged = false;
            }
        } 
        catch (error) {
            this.showErrors([error.message]);
        }
        Field.prototype.onInputFocusout.call(this, $$input, event);
    },
    onInputChange: function($$input, event){
        try {
            // handle formatting
            var self = this, val = $$input.val(), valFormatted = val;
            self.valHasChanged = true;
            if (val != '') {
                $$input.val((valFormatted = this._formatInput(self, val)));
            }
            if (self.validate(valFormatted, [])) {
                self.onInputValidate(event);
            }
        } 
        catch (error) {
            this.showErrors([error.message]);
        }
    },
    onInputValidate: function(event){
        try {
            var newValue = this.getDataValue();
            if (newValue) {
                this.notifyFieldChange(newValue);
            }
        } 
        catch (error) {
            this.showErrors([error.message]);
        }
    },
    onInputKeypress: function($$input, event){
        /*
         var self = this;
         var type = self.$field.$type;
         
         // === building regexp ===
         var pattern = "0-9+-";
         var decimal = ".,";
         var regex;
         pattern += type == "application/x-decimal" ? decimal : '';
         regex = new RegExp("[" + pattern + "]");
         */
        /*
         // === blocking alphabetic characters ===
         var key = event.which;
         var c = String.fromCharCode(key);
         if (event.ctrlKey || event.altKey || event.metaKey || key < 32) {
         return true;
         }
         if (/[a-zA-Z]/g.test(c)) {
         event.preventDefault();
         return false;
         }*/
        var self = this;
        // format properties
        var numFormatObj = self.numFormatObj;
        var gs = numFormatObj.groupSeparator;
        
        // -- building regex pattern --
        var chars_ok = "0-9-+"; // digits and signs
        chars_ok += self.$field.$type == "application/x-integer" ? '' : numFormatObj.decimalSeparator;
        chars_ok += gs == " " ? "\\s" : gs;
        
        var regex = new RegExp("[" + chars_ok + "]");
        
        // -- blocking forbidden characters --
        var key = event.which;
        var c = String.fromCharCode(key);
        if (event.ctrlKey || event.altKey || event.metaKey || key < 32) {
            return true;
        }
        if (!regex.test(c)) {
            // error color
            $$input.toggleClass('s-field-input-error', true);
            event.preventDefault();
            return false;
        }
        else {
            $$input.toggleClass('s-field-input-error', false);
        }
    },
    onInputKeydown: function($$input, event){
        // handling escape button
        /*
         var key = event.which;
         var self = this;
         var $$input = self.$$input;
         if (key == '27') {
         var numObj = self.formatApi.parse($$input.val());
         $$input.val(self.formatApi.format(numObj, self.localFormat));
         event.preventDefault();
         return false;
         }
         */
    },
    _doClickPicker: function(btn){
        if (!this.$isDisabled && !this.$isReadOnly) {
            var step = this.formatApi.parse(this.$item.$step || "1"); // assuming $item.$step may be set in field object
            var numObj = this.formatApi.parse(this.$$input.val());
            this.currentValue = this.formatApi.numberValue(this.formatApi[(btn.getAttribute("data-s-picker") == "numeric-up") ? "add" : "subtract"](numObj, step));
            this.$$input.val(this.currentValue).focus().change();
        }
    },
    render: function(){
        if (this.$isEditMode) {
            this._input = document.createElement("input");
            this._input.type = "text";
            this.$$input = $(this._input);
            this._input.className = this.$skinInput + " " + this.$skinInput + "-num";
            this.fieldValue.appendChild(this._input);
            this._appendLinksPicker();
            return true;
        }
        return false;
    },
    setDataValue: function(value, record){
        try {
            var formattedNum = '';
            this.currentValue = value;
            if (value) {
                // building formatted value
                var numObj = this.formatApi.parse(value);
                formattedNum = this.formatApi.format(numObj, this.localFormat);
            }
            if (this.$isEditMode) {
                this.$$input.val(formattedNum);
            }
            else {
                document.site.emptyDom(this.fieldValue);
                if (formattedNum && this.$field.$links && this.$field.$links.$details && !this.$item.$isMenusDisabled) {
                    this.appendDetailLink(formattedNum, this.$field.$links.$details);
                }
                else {
                    this.fieldValue.textContent = formattedNum;
                }
            }
        } 
        catch (error) {
            this.showErrors([error.message]);
        }
    },
    getDataValue: function(){
        return this.formatApi.parse(this.$$input.val(), this.localFormat).toString();
    }
});
