// syracuse-ui/lib/field/phoneField.js
"use strict";
var _helpers = require('syracuse-core/lib/helpers');
var Field = require('syracuse-ui/lib/field/field').Field;

exports.PhoneField = _helpers.defineClass(function() {}, Field, {
	initialize: function() {
		this.$inputType = "tel";
		this.$isDetailLinkDisabled = true;
	},
	_createValueContainer: function() {
		if (!this.$isEditMode) {
			this._dataValue = document.createElement("a");
		}
		Field.prototype._createValueContainer.call(this);
		if (!this.$isEditMode) {
			this._dataValue.className = this._$cssField + "-value-read s-field-phone";
		}
	},
	setDataValue: function(value) {
		this.currentValue = value;
		if (this.$isEditMode) {
			this.setInputValue(this.currentValue || "");
		} else {
			this.displayText = value;
			if (value) {
				this._dataValue.syraIsDefaultLink = true;
				this._dataValue.href = "tel:" + value;
				this._dataValue.textContent = value;
				if (!this._iconFlag) {
					this._iconFlag = syra_menus.button.add({
						isIndicator: true,
						parent: this,
						iconOnly: true,
						css: " s-icon-phone",
						fontIcon: "field_phone"
					});
				}!this._iconFlag.link.parentNode && this._dataValue.insertBefore(this._iconFlag.link, this._dataValue.firstChild);
			} else {
				syra_menus.button.remove(this._iconFlag);
				this._dataValue.href = "";
				this._dataValue.textContent = value;
			}
			syra_menus.pickers.menus.show(this);
		}
	}
});
