// syracuse-ui/lib/field/textPlainField.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Field = require('syracuse-ui/lib/field/field').Field;
var locale = require('syracuse-core/lib/locale');

function TextPlainField() {}

exports.TextPlainField = helpers.defineClass(TextPlainField, Field, {
	initialize: function() {
		this.$item.$rows = this.$item.$rows || 5;
		this.$lazyLoading = this.$field.$url != null;
	},
	setDataValue: function(value) {
		var self = this;
		if (this.$lazyLoading && value && typeof(value) === 'object') {
			// lazy loading
			self.$url = this.articleParent.parseExpression(value.$url || this.$field.$url);
			self.$contentType = value.$contentType || this.$field.$contentType;
			self.$url && document.controller.sendRequest(null, {
				$location: {
					$url: self.$url,
					$contentType: self.$contentType,
					$type: self.$contentType
				}
			}, function(data, response) {
				if (typeof(data) === 'string') {
					Field.prototype.setDataValue.call(self, data);
				} else {
                    self.showErrors([locale.format(self.getLocalize().f_badDataType, "string", typeof(data))]);
				}
			}, function(error, httpquery) {
				self.showErrors([error.message]);
			});
		} else {
			Field.prototype.setDataValue.call(self, value);
		}
	},
	notifyFieldChange: function(newValue, validated) {
		var self = this;
		if (this.$lazyLoading && this.$url) {
			document.controller.sendRequest(null, {
				$location: {
					$url: self.$url,
					$contentType: self.$contentType,
					$type: self.$contentType
				},
				data: newValue,
				method: "PUT"
			}, function(data, response, requestUrl) {});
		} else {
			Field.prototype.notifyFieldChange(this, newValue, validated);
		}
	}
});
