// syracuse-ui/lib/field/unknowField.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Field = require('syracuse-ui/lib/field/field').Field;

function UnknowField() {}

exports.UnknowField = helpers.defineClass(UnknowField, Field, {
	applyDesignMetaData: function(metaData, onAuthoring) {},
	onInputFocusin: function(input, event) {},
	onInputClick: function(input, event) {},
	onInputFocusout: function(input, event) {},
	appendDetailLink: function(value, $detailsLink) {},
	hideTitle: function($isTitleHidden) {},

	loadBox: function($$item) {
		this.$isEditMode = false;
		this.boxPickersPadding = 0;
		this.articleParent = this.boxParent.getArticle();
		this.ensureEditMode();
		this.$skinField = this.$skin = this.$item.$skin || (this.$item.$inplace ? "s-inplace" : "s-field");
		if (this.articleParent.isRecordArticle) {
			this.arrayLevel = "cell";
		}
		this._initCss();
		this._$cssTopField = "s-field-unknow " + this._$cssTopField;
		this._$cssField = "s-field-unknow " + this._$cssField;

		this.$$item = $$item || $(this.layoutSlot);
		if (!this.$item.$inplace) {
			this.$$item = $$item || $(document.createElement("div")).appendTo(this.$$item);
		}
		(this._domItem = this.$$item[0]).className += " " + this._$cssTopField;

		this.$rootLinks = this.$field;

		this.appendTitle();

		this.domValueSlot = document.createElement("div");
		this.domValueSlot.className = this._$cssField + "-value";
		this.$$dataValue = $(this._dataValue = this.domValueSlot);
		if (this.$item.$inplace) {
			this._core = this.domValueSlot;
			this._domItem.appendChild(this.domValueSlot);
		} else {
			this.domValueSlot = document.createElement("div");
			this.domValueSlot.className = this._$cssField + "-value-slot";
			this.domValueSlot.appendChild(this._dataValue);

			this._core = document.createElement("div");
			this._core.className = this._$cssTopField + "-core";
			this._core.appendChild(this.domValueSlot);
			this._domItem.appendChild(this._core);
		}
		this._dataValue.appendChild(this.fieldValue = document.createElement("div"));
		this.fieldValue.className = this._$cssField + "-value-read";

		this.setTitle(this.$item.$title || this.$field.$title);

		this.fieldValue.textContent = "Unknown: " + this.$item.$bind + "  type=" + (this.$field.$type || "");
		this._domItem.setAttribute("data-s-field", this.id);
		this._domItem.setAttribute("data-s-field-item", this.$item.$bind || "");
		if (!this.$item.$inplace && !this.$isEditMode) {
			this.domValueSlot.style.display = 'inline-block';
		}
		this.ensureLayoutMode();
	},
	setFieldWidth: function() {},
	applyFieldStyle: function(metaData) {},
	applyAutoSize: function(state) {},
	_bindAutoSize: function(bind) {},
	render: function() {

	},
	_updateBoxPickerPadding: function(add) {},
	createPicker: function(pickerType) {
		return btn;
	},
	setDescription: function($description) {

	},
	setState: function(state) {

	},
	_onSetReadOnly: function($isReadOnly) {

	},
	setDataBind: function(value, record, metaData, $bind) {

	},
	setDataValue: function(value, $bind) {

	},
	saveClientValue: function() {

	},
	getInputValue: function() {
		return "";
	},
	setInputValue: function(val) {

	},
	applyMetaData: function(metaData, options) {

	},
	focus: function(select) {
		return false;
	},
	toggleDiagnose: function(css, show) {

	},
	drawDiagnose: function($diagnose, severityGroup) {

	},
	emptyDiagnoseSlot: function() {

	},
	ensureDiagnoseSlot: function() {

	},
	showErrors: function(errors, options) {

	},
	showDiagnoses: function(diagnoses, options) {

	},
	validate: function(newValue, errors) {
		return false;
	},
	onInputValidate: function(event, notify) {

	},
	onInputChange: function(input, event) {

	},
	onInputKeyup: function(input, event) {},
	notifyFieldChange: function(newValue, validated) {

	},
	toggleEditMode: function($isEditMode) {

	},
	setMenus: function(metaData) {

	},
	setMenusBox: function(metaData) {},
	filterMenu: function($bind) {
		return true;
	},
	_createMenusBox: function() {},
	_isShowMenuPicker: function() {},
	_ensureMenuPickerVisible: function(isRemoved) {

	},
	ensureLayoutMode: function() {}
});
