// syracuse-ui/lib/field/unknowField.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Field = require('syracuse-ui/lib/field/field').Field;

function UnknowField() {}

exports.UnknowField = helpers.defineClass(UnknowField, Field, {
	applyDesignMetaData: function(metaData, designing) {},
	onInputFocusin: function(input, event) {},
	onInputClick: function(input, event) {},
	onInputFocusout: function(input, event) {},
	hideTitle: function($isTitleHidden) {},
	dispose: function() {
		if (this.$item) {
			var unknowFields = this.articleParent && this.articleParent.unknowFields;
			if (unknowFields) {
				delete unknowFields[this.$item.$bind];
			}
		}
		Field.prototype.dispose.call(this);
	},
	loadBox: function() {
		(this.articleParent.unknowFields = this.articleParent.unknowFields || {})[this.$item.$bind] = this;
		this.isDiagnoseDisabled = true;
		this.$isEditMode = false;
		this.$designLevel = "field";
		this.ensureEditMode();
		this.$skinField = this.$skin = this.$item.$skin || (this.$item.$inplace ? "s-inplace" : "s-field");
		if (this.articleParent.isRecordArticle) {
			this.arrayLevel = "cell";
		}
		this._initCss();
		var css = "s-field-" + this.unknowMode;
		this._$cssField = css + " " + this._$cssField;

		this.domItem = this.layoutSlot;
		if (!this.$item.$inplace) {
			this.domItem = this.layoutSlot.appendChild(document.createElement("div"));
		}
		this.domItem.className += " " + this._$cssField;

		this.appendTitle();
		this.domTitle && syra_dom.hide(this.domTitle, true);
		this._createValueContainer();

		this.setTitle(this.$item.$title || this.$field.$title);

		switch (this.unknowMode) {
			case "nowidget":
				this._dataValue.textContent = syra_local.fieldUnknowType + ": " + this.$item.$bind + "  type=" + (this.$field.$type || "");
				break;
			case "noproto":
				this.isItemHidden = true;
				this._dataValue.textContent = syra_local.fieldNoProtoype + ": " + this.$item.$bind;
				(this.page.noProtoFields = this.page.noProtoFields || []).push(this);
				break;

		}
		this.isField = true;
		this.domItem.syrainout = this.domItem.syraItem = this.id;
		this.domItem.syra_field_bind = this.$item.$bind || "";
		if (!this.$item.$inplace && !this.$isEditMode && !this.$item.$isAdvanced) {
			this._dataValue.style.display = 'inline-block';
		}
		syra_fields.advancedState.setState(this);
		this.ensureLayoutMode();
		if (this.unknowMode == "noproto") {
			this.showItem(false);
		}
	},
	applyAutoSize: function(state) {},
	render: function() {

	},
	setDescription: function($description) {

	},
	setState: function(state) {

	},
	setDataBind: function(value, record, metaData, $bind) {

	},
	setDataValue: function(value, $bind) {

	},
	getDatasetValue: function() {},
	setDatasetValue: function() {

	},
	getInputValue: function() {
		return "";
	},
	setInputValue: function(val) {

	},
	applyMetaData: function(metaData, options) {

	},
	focus: function(select) {
		return false;
	},
	showErrors: function(errors, options) {

	}
});
