// syracuse-ui/lib/field/urlField.js
"use strict";
var _helpers = require('syracuse-core').helpers;
var Field = require('syracuse-ui/lib/field/field').Field;

exports.UrlField = _helpers.defineClass(function() {}, Field, {
	initialize: function() {
		this.$inputType = "url";
		this.$isDetailLinkDisabled = true;
	},
	_createValueContainer: function() {
		if (!this.$isEditMode) {
			this._dataValue = document.createElement("a");
		}
		Field.prototype._createValueContainer.call(this);
		if (!this.$isEditMode) {
			this._dataValue.className = this._$cssField + "-value-read s-mn-link";
		}
	},
	setDataValue: function(value) {
		this.currentValue = value;
		if (this.$isEditMode) {
			this.setInputValue(this.currentValue || "");
		} else {
			if (value) {
				this._dataValue.setAttribute("href", value);
				this._dataValue.textContent = this.displayText = value;
				this._dataValue.syraOnClick = "onUrlClick";
				if (this._core != this._dataValue) {
					this._core.appendChild(this._dataValue);
				}
			} else {
				this._dataValue.parentNode && this._dataValue.parentNode.removeChild(this._dataValue);
			}
			syra_menus.pickers.menus.show(this);
		}

	},
	onUrlClick: function() {
		this.currentValue &&
			syra_url.history.load({
				$url: this.currentValue,
				target: "blank",
				isDefaultUrl: false
			});
	}

});
