// syracuse-tablet/html/js/old/demos/_entities/entTodos.js
"use strict";

var _todos = [{
	$uuid: "1",
	date: "2015-02-01 09:00:00",
	who: "Mr. Miller",
	desc: "Call",
	phone: "+771 896520 4183"
}, {
	$uuid: "2",
	date: "2015-02-09 10:00:00",
	who: "Mr. Miller",
	desc: "Meeting"
}, {
	$uuid: "3",
	date: "2015-03-04 09:00:00",
	who: "Mr. Heinz",
	desc: "Milestone definition"
}, {
	$uuid: "4",
	date: "2015-03-05 15:00:00",
	who: "Mr. Schmidt",
	desc: "Project review",
	phone: "+771 896520 4183"
}, {
	$uuid: "5",
	date: "2015-03-08 14:00:00",
	who: "Mr. Miller",
	desc: "Call"
}, {
	$uuid: "6",
	date: "2015-03-09 15:40:00",
	who: "Mr. Miller",
	desc: "Call"
}, {
	$uuid: "7",
	date: "2015-03-10 15:50:00",
	who: "Mr. Miller",
	desc: "Call"
}, {
	$uuid: "8",
	date: "2015-03-11 14:00:00",
	who: "Mr. Miller",
	desc: "Call"
}, {
	$uuid: "9",
	date: "2015-03-11 14:00:00",
	who: "Mr. Miller",
	desc: "Call"
}, {
	$uuid: "10",
	date: "2015-03-11 14:00:00",
	who: "Mr. Miller",
	desc: "Call"
}, {
	$uuid: "11",
	date: "2015-03-11 14:00:00",
	who: "Mr. Miller",
	desc: "Call"
}, {
	$uuid: "12",
	date: "2015-03-11 14:00:00",
	who: "Mr. Miller",
	desc: "Call"
}, {
	$uuid: "13",
	date: "2015-03-11 14:00:00",
	who: "Mr. Miller",
	desc: "Call",
}];

exports.entity = {
	$entityName: "$localTodo",
	$value: "{desc}",
	$properties: {
		who: {
			$type: "application/x-string",
			$title: "Who"
		},
		desc: {
			$type: "application/x-string",
			$title: "Description"
		},
		phone: {
			$type: "application/x-string",
			$title: "Phone",
			$format: "$phone"
		},
		date: {
			$type: "application/x-datetime",
			$title: "Date"
		},
	},
	$articles: {},
	$services: {
		$new: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve();
			return deferred.promise();
		},
		$read: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve(_todos[op.$key]);
			return deferred.promise();
		},
		$query: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve({
				$resources: _todos
			});
			return deferred.promise();
		},
		$save: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve();
			return deferred.promise();
		},
		$delete: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve();
			return deferred.promise();
		}
	}
};
