// @sage/syracuse-react/src/test/over/overTest.tsx
"use strict";
import * as React from 'react';
import Section from '../../section/section';
import { Link } from '../../link/link';
import { lock, enableLong } from '../../over/lock/lockController';
import * as Formula from './data/formula';
import * as PrinterSetup from './data/printerSetup';
import * as SyraPageModal from '../../over/modal/page/syraPageModal';

export default class OverTest extends React.Component<{}, {}>{
    static representationName: string = "over";

    onFunctionalClick(value: any) {
        let syra_site = (window as any).syra_site;
        let $prototype = syra_site.functionalTools[value].getPage();
        SyraPageModal.loadModal({
            $facet: "edit",
            initData: value == "formula" ? Formula.data : PrinterSetup.data,
            $representation: {
                $prototype: $prototype,
                $article: $prototype.$article
            }
        }, {
                authoringDisabled: true
            });
    }
    onLockClick = (long: boolean) => {
        lock();
        long && enableLong();
    }
    render() {
        return <div>
            <Section title="Modals">
                <Link
                    className="s_textLink"
                    value={"formula"}
                    title="formula"
                    onClick={this.onFunctionalClick} />
                <Link
                    className="s_textLink"
                    value={"printerSetup"}
                    title="printerSetup"
                    onClick={this.onFunctionalClick} />
            </Section>
            <Section title="Lock">
                <Link
                    className="s_textLink"
                    title="ui lock"
                    value={false}
                    onClick={this.onLockClick} />
                <Link
                    className="s_textLink"
                    title="ui lock long"
                    value={true}
                    onClick={this.onLockClick} />
            </Section>
        </div>;
    }
}
