// etna/lib/supervisor/meta/collection.ts
import { _ } from 'streamline-runtime';
import * as types from '../../engine/runtime/types';
import * as util from '../util';
import { BaseClass } from './baseClass';
import { ColumnDesc } from './column';
import { Property } from './property';
import { Supervisor } from '../supervisor';
const tracerJs = require('syracuse-core').getTracer("etna.supervisor");

export interface ClassLinkMap {
	PROMAP: string;
	KEYMAP: string;
	VALDEFMAP: string;
}

export interface ClassLinkData {
	TYPLNK: number;
	MAPS: ClassLinkMap[];
	CLALNK: string;
	CLELNK: string;
	EXPLNK: string; 
	TABLNK: string;
	[name: string]: any;
}

export interface CollectionData {
	CODCOL: string;
	PROCOL: string;
	INTCOL: number;
	MAXCOL: number;
	MINCOL: number;
	FLGAPDCOL: number;
	FLGINSCOL: number;
	FLGSUPCOL: number;
	FLGTRICOL: number;
	ACTGRP: string;
	_proxyName: string;
}

export class Collection {
	supervisor: Supervisor;
	data: CollectionData;
	class: BaseClass;
	property: Property;
	mappingData: ClassLinkData;
	ofIndexedProperties: boolean;
	count: number;
	constructor(superv: Supervisor, clas: BaseClass, data: CollectionData) {
		this.supervisor = superv;
		this.class = clas;
		this.data = data;
	}
	init(_: _) {
		return this;
	}
	get name() {
		return this.data.CODCOL;
	}
	get jsName() {
		return this.data._proxyName;
	}
	get isInstance() {
		return this.mappingData && this.mappingData.TYPLNK === 3;
	}
	get nbrPro() {
		return this.class.data.PROPERTIES.length;
	}
	getSql(_: _, colNames: string[], tableNames: string[], wheres: string[], params: string[], descs: ColumnDesc[], 
		cache: any, raw: boolean, sqlOpts?: types.DriverReaderOptions) {
		const properties = this.class.properties;
		tracerJs.debug && tracerJs.debug("collection.getSql:" + Object.keys(properties));
		this.class.cache = cache;
		Object.keys(properties).forEach_(_, (_, name) => {
			properties[name].getSql && properties[name].getSql(_, colNames, tableNames, wheres, params, descs, cache, raw, sqlOpts);
		}, this);
	}
};

export const Constructor = Collection;
