// etna/lib/supervisor/meta/collection.ts
import { _ } from 'streamline-runtime';
import { DriverReaderOptions } from '../../engine/drivers/types';
import * as types from '../../engine/runtime/types';
import * as util from '../util';
import { BaseClass } from './baseClass';
import { ColumnDesc } from './column';
import { Property } from './property';
import { Supervisor } from '../supervisor';
import { CollectionMeta, ClassLinkMeta } from '../metadata/class';
const tracerJs = require('syracuse-core').getTracer("etna.supervisor");

export class Collection {
	supervisor: Supervisor;
	data: CollectionMeta;
	class: BaseClass;
	property: Property;
	mappingData: ClassLinkMeta;
	ofIndexedProperties: boolean;
	count: number;
	constructor(superv: Supervisor, clas: BaseClass, data: CollectionMeta) {
		this.supervisor = superv;
		this.class = clas;
		this.data = data;
	}
	init(_: _) {
		return this;
	}
	get name() {
		return this.data.CODCOL;
	}
	get jsName() {
		return this.data._proxyName;
	}
	get isInstance() {
		return this.mappingData && this.mappingData.TYPLNK === 3;
	}
	get nbrPro() {
		return this.class.data.PROPERTIES.length;
	}
	getSql(_: _, colNames: string[], tableNames: string[], wheres: string[], params: string[], descs: ColumnDesc[], 
		cache: any, raw: boolean, sqlOpts?: DriverReaderOptions) {
		const properties = this.class.properties;
		tracerJs.debug && tracerJs.debug("collection.getSql:" + Object.keys(properties));
		this.class.cache = cache;
		Object.keys(properties).forEach_(_, (_, name) => {
			properties[name].getSql && properties[name].getSql(_, colNames, tableNames, wheres, params, descs, cache, raw, sqlOpts);
		}, this);
	}
};

export const Constructor = Collection;
