// etna/lib/supervisor/meta/collection.ts
import { _ } from 'streamline-runtime';
import * as util from '../util';
import { BaseClass } from './baseClass';
import { Supervisor } from '../supervisor';
const tracerJs = require('syracuse-core').getTracer("etna.supervisor");

export interface CollectionMappingData {
	TYPLNK: number;
}

export interface CollectionData {
	CODCOL: string;
	_proxyName: string;
}

export class Collection {
	supervisor: Supervisor;
	data: CollectionData;
	class: BaseClass;
	mappingData: CollectionMappingData;
	constructor(superv: Supervisor, clas: BaseClass, data: CollectionData) {
		this.supervisor = superv;
		this.class = clas;
		this.data = data;
	}
	init(_: _) {
		return this;
	}
	get name() {
		return this.data.CODCOL;
	}
	get jsName() {
		return this.data._proxyName;
	}
	get isInstance() {
		return this.mappingData && this.mappingData.TYPLNK === 3;
	}
	get nbrPro() {
		return this.class.data.PROPERTIES.length;
	}
	getSql(_: _, colNames: string[], tableNames: string[], wheres, params, descs, cache, raw, sqlOpts) {
		const properties = this.class.properties;

		tracerJs.debug && tracerJs.debug("collection.getSql:" + Object.keys(properties));
		this.class.cache = cache;
		Object.keys(properties).forEach_(_, (_, name) => {
			properties[name].getSql && properties[name].getSql(_, colNames, tableNames, wheres, params, descs, cache, raw, sqlOpts);
		}, this);
	}
};

export const Constructor = Collection;
