// syracuse-ui/lib/field/vignette/richData/richDataVignetteField.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var VignetteField = require("syracuse-ui/lib/field/vignette/vignetteField").VignetteField;

var _viewerTypes = {
	"text/xml": require('./viewer/xmlViewer').XMLViewer,
	"application/json": require('./viewer/jsonViewer').JSONViewer
};

function RichDataVignetteField() {}

exports.RichDataVignetteField = helpers.defineClass(RichDataVignetteField, VignetteField, {
	addRefreshButton: function() {
		//avoid display of refresh button
	},
	setDataBind: function(value, record, metaData, $bind) {
		if (metaData && metaData.$title !== undefined) {
			this.setTitle(metaData.$title);
		}
		this.currentValue = value;
		this.renderData();
	},
	resizeItem: function(doResize) {
		if (this.$item.$height) {
			this._dataValue.style.height = this.$item.$height + "px";
		} else {
			this._dataValue.style.height = this.page.size.height * 0.3 + "px";
		}
	},
	loadBox: function() {
		VignetteField.prototype.loadBox.call(this);
		this._dataValue = syra_dom.addDiv("s-richdata-viewer", this.body);
	},
	renderData: function() {
		if (this.currentValue) {
			this._dataValue.style.height = this.page.scrollview.clientHeight * 0.3 + "px";
			this.discoverViewer();
		} else {
			syra_dom.empty(this._dataValue);
			this.viewer && this.viewer.dispose();
			this.viewer = null;
		}
	},
	tryViewer: function(type) {
		this.viewer = new _viewerTypes[type]();
		return this.viewer.render(this._dataValue, this.currentValue);
	},
	discoverViewer: function() {
		var dataset = this.articleParent.dataset;
		var $field = dataset && dataset.$properties && dataset.$properties[this.$item.$bind];
		var $type = ($field && $field.$type) || this.$field.$type;
		syra_dom.empty(this._dataValue);
		syra_dom.toggleClass(this._dataValue, "s-error", false);
		//try defined type 
		var error = this.tryViewer($type);
		if (error) {
			var keys = Object.keys(_viewerTypes);
			for (var ii = 0, jj = keys.length; ii < jj; ii++) {
				var type = keys[ii];
				if (type != $type) {
					if (!this.tryViewer(type)) {
						return; //no error returned,render with success
					}
				}
			}
			syra_dom.toggleClass(this._dataValue, "s-error", true);
			this._dataValue.textContent = error;
		}
	},
	dispose: function() {
		this.renderData(null);
		VignetteField.prototype.dispose.call(this);
	}
});
