// syracuse-tablet/html/js/storage/storageWebSQLDrafts.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var storageInterface = require('syracuse-tablet/html/js/storage/storageInterface');
var Base = require('syracuse-tablet/html/js/storage/storageWebSQLSData').Klass;
var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("drafts", false);
/**
 * WebSQL storage interface - DRAFTS MANAGEMENT
 * Inherits from storageWebSQLSData only to split the code
 */

var _DRAFT_INSERT = "INSERT INTO sdata_drafts (id, endpoint, representation, comment, status, errorMsg, dataSet, draftUrl, title, creation_date) values (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)";
var _DRAFT_DELETE_ALL = "DELETE FROM sdata_drafts";
var _DRAFT_DELETE = "DELETE FROM sdata_drafts where id = ?";
var _DRAFT_READ = "SELECT * FROM sdata_drafts where id = ?";
var _DRAFT_READ_LIST = "SELECT id, endpoint, representation, comment, status, errorMsg, draftUrl, title, creation_date FROM sdata_drafts";
var _DRAFT_COUNT = "SELECT COUNT(*) AS counter FROM sdata_drafts";
var _tables = {
	"sdata_drafts": [
		"DROP TABLE IF EXISTS sdata_drafts",
		"CREATE TABLE sdata_drafts ( " +
		"id text, " +
		"endpoint text, " +
		"representation text, " +
		"comment text, " +
		"status text, " +
		"errorMsg text, " +
		"dataSet text, " +
		"draftUrl text, " +
		"title text, " +
		"creation_date text" +
		")",
		"DROP INDEX IF EXISTS sdata_drafts_id",
		"CREATE UNIQUE INDEX sdata_drafts_id ON sdata_drafts (id)"
	]
};

var _logError = function(title, status) {
	log && log(title + " ERROR");
	log && log(storageInterface.resultToString(status));
};
/**
 * Manages drafts and working copies storage
 */
exports.Klass = utils.defineClass(
	function() {
		Base.call(this);
	}, Base, {
		getTables: function() {
			var res = Base.prototype.getTables.call(this);
			$.extend(res, _tables);
			return res;
		},
		_draftReadInfo: function(item) {
			if (!item) return null;
			try {
				return {
					id: item.id,
					endpoint: item.endpoint || "",
					representation: item.representation || "",
					comment: item.comment || "",
					status: item.status || "none",
					errorMsg: item.errorMsg || "",
					dataSet: item.dataSet ? JSON.parse(item.dataSet) : null,
					draftUrl: item.draftUrl || "",
					title: item.title || "",
					creation_date: item.creation_date
				};
			} catch (e) {
				log && log("_draftReadInfo", "KO", item.id);
				return null;
			}
		},
		/**
		 * Resolve with [resources]/[]
		 */
		_draftReadList: function() {
			var self = this;
			return self._executeSql(_DRAFT_READ_LIST)
				.then(function(result) {
					if (!result || !result.rows || result.rows.length == 0) return [];
					var resources = [],
						rec;
					for (var i = 0; i < result.rows.length; i++) {
						rec = self._draftReadInfo(result.rows.item(i));
						if (rec) {
							resources.push(rec);
						}
					}
					log && log("_draftReadList", "OK");
					return resources;
				}).fail(function(status) {
					_logError("_draftReadList", status);
					return [];
				});
		},
		/**
		 * Resolve with data/null
		 */
		_draftRead: function(id) {
			var self = this;
			return self._executeSql(_DRAFT_READ, [id])
				.then(function(result) {
					if (!result || !result.rows || result.rows.length == 0) return null;
					log && log("_draftRead", id, "OK");
					return self._draftReadInfo(result.rows.item(0));
				}).fail(function(status) {
					_logError("_draftRead", status);
					return null;
				});
		},
		/**
		 * Resolve with  true/false
		 */
		_draftDelete: function(id) {
			var self = this;
			return self._executeSql(_DRAFT_DELETE, [id]).then(function(result) {
				log && log("_draftDelete", id, result ? "OK" : "KO");
				return true;
			}).fail(function(status) {
				_logError("_draftDelete", status);
				return false;
			});
		},
		/**
		 * Resolve with true/false
		 */
		_draftDeleteAll: function() {
			var self = this;
			return self._executeSql(_DRAFT_DELETE_ALL).then(function(result) {
				log && log("_draftDeleteAll OK");
				return true;
			}).fail(function(status) {
				_logError("_draftDeleteAll", status);
				return false;
			});
		},
		/**
		 * Resolve with draft context or reject with an exception
		 */
		_draftSave: function(ctx) {
			var self = this;
			var deferred = $.Deferred();
			if (!ctx || !ctx.id) {
				log && log("_draftSave error", "Invalid parameters");
				deferred.reject(new Error("Invalid parameters"));
			} else {
				self._draftDelete(ctx.id).then(function() {
					return self._executeSql(_DRAFT_INSERT, [ctx.id, ctx.endpoint, ctx.representation, ctx.comment, ctx.status, ctx.errorMsg, JSON.stringify(ctx.dataSet), ctx.draftUrl, ctx.title, ctx.creation_date]);
				}).then(function() {
					log && log("_draftSave", ctx.id, "OK");
					deferred.resolve(ctx);
				}).fail(function(status) {
					_logError("_draftSave", status);
					deferred.reject(status.$exception || new Error(status.message));
				});
			}
			return deferred.promise();
		},
		_draftCount: function() {
			return this._executeSql(_DRAFT_COUNT).then(function(result) {
				return result && result.rows.length > 0 ? result.rows.item(0).counter : 0;
			}).fail(function(status) {
				_logError("_draftCount", status);
				return 0;
			});
		},
		/**
		 * Drafts
		 * op: Operation
		 * 		save
		 * 			args(1) create true/false
		 * 		read
		 * 		query
		 * 		delete
		 * 		switchStatus
		 */
		draftOperation: function(op) {
			switch (op) {
				case "draftSave":
					return this._draftSave(arguments[1]);
				case "draftDelete":
					return this._draftDelete(arguments[1]);
				case "draftDeleteAll":
					return this._draftDeleteAll();
				case "draftRead":
					return this._draftRead(arguments[1]);
				case "draftReadList":
					return this._draftReadList();
				case "draftCount":
					return this._draftCount(false);
				default:
					return $.smReject(storageInterface.buildErrResult("Unknown Draft operation[" + op + "]"));
			}
		}
	});
