// syracuse-tablet/html/js/storage/storageWinJS.js
"use strict";

/*
 * WinJS storage interface (Webview control running in Windows Store Application)
 */

var utils = require('syracuse-tablet/html/js/helpers/utils');
var storageInterface = require('syracuse-tablet/html/js/storage/storageInterface');
var winjs = require('syracuse-tablet/html/js/helpers/winjs');

function _StorageInterfaceWinJS() {
	this.infoString = "WinJS storage interface";
};

/*
 *
 * Proxy class to route all call to winJS container
 *
 */
var _storageInterfaceWinJSClass = utils.defineClass(
	_StorageInterfaceWinJS,
	storageInterface.StorageInterface, {
		init: function(useTestDatabase) {
			return winjs.callWinJS("winjsDB", "init", {
				useTestDatabase: useTestDatabase
			});
		},

		queryAppConfigs: function() {
			return winjs.callWinJS("winjsDB", "queryAppConfigs", {});
		},

		createAppConfig: function(data) {
			return winjs.callWinJS("winjsDB", "createAppConfig", {
				app: data
			});
		},

		readAppConfig: function(appid) {
			return winjs.callWinJS("winjsDB", "readAppConfig", {
				appid: appid
			});
		},

		deleteAppConfig: function(appid) {
			return winjs.callWinJS("winjsDB", "deleteAppConfig", {
				appid: appid
			});
		}
	}
);

exports.StorageInterface = _storageInterfaceWinJSClass;
exports.isAvailable = function() {
	return winjs.isAvailable();
};
