// syracuse-tablet/html/js/storage/storageWebSQL.js
"use strict";

/*
 * WebSQL storage interface (Chrome, Safari)
 */

var utils = require('syracuse-tablet/html/js/helpers/utils');
var storageInterface = require('syracuse-tablet/html/js/storage/storageInterface');

function _StorageInterfaceWebSQL() {
	this.infoString = "WebSQL storage interface";
};

var _storageInterfaceWebSQLClass = utils.defineClass(
	_StorageInterfaceWebSQL,
	storageInterface.StorageInterface, {
		init: function(cb, useTestDatabase) {
			var deferred = $.Deferred();
			try {
				deferred.resolve();
			} catch (e) {
				deferred.reject();
			} finally {
				return deferred.promise();
			}
		},

		/*==================================
         * App management
         ==================================*/
		queryAppConfigs: function() {},

		//was: installApp
		createAppConfig: function(data) {},

		//getAppConfig
		readAppConfig: function(appid) {},

		// was: removeFromDB
		deleteAppConfig: function(id) {},

	}
);

exports.StorageInterface = _storageInterfaceWebSQLClass;
exports.isAvailable = function() {
	return (window.openDatabase != null);
};
