// syracuse-tablet/html/js/storage/storageWebSQLTemplates.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var storageInterface = require('syracuse-tablet/html/js/storage/storageInterface');
var Base = require('syracuse-tablet/html/js/storage/storageWebSQLDrafts').Klass;
var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("templates", true);
/**
 * WebSQL storage interface - TEMPLATES MANAGEMENT
 * Inherits from storageWebSQLSData only to split the code
 */

var _TMPL_INSERT = "INSERT INTO sdata_templates (id, endpoint, representation, title, dataSet, creation_date) values (?, ?, ?, ?, ?, ?)";
var _TMPL_DELETE_ALL = "DELETE FROM sdata_templates";
var _TMPL_DELETE = "DELETE FROM sdata_templates where id = ?";
var _TMPL_READ = "SELECT * FROM sdata_templates where id = ?";
var _TMPL_QUERY = "SELECT id, endpoint, representation, title, creation_date FROM sdata_templates";
var _TMPL_COUNT = "SELECT COUNT(*) AS counter FROM sdata_templates";
var _tables = {
	"sdata_templates": [
		"DROP TABLE IF EXISTS sdata_templates",
		"CREATE TABLE sdata_templates ( " +
		"id text, " +
		"endpoint text, " +
		"representation text, " +
		"title text, " +
		"dataSet text, " +
		"creation_date text" +
		")",
		"DROP INDEX IF EXISTS sdata_templates",
		"CREATE UNIQUE INDEX sdata_templates_id ON sdata_templates (id)"
	]
};

var _logError = function(title, status) {
	log && log(title + " ERROR");
	log && log(storageInterface.resultToString(status));
};
/**
 * Manages templates (edit/create facets) storage
 */
var _Klass = utils.defineClass(
	function() {
		Base.call(this);
	}, Base, {
		getTables: function() {
			var res = Base.prototype.getTables.call(this);
			$.extend(res, _tables);
			return res;
		},
		_tmplReadInfo: function(item) {
			if (!item) return null;
			try {
				return {
					id: item.id,
					endpoint: item.endpoint || "",
					representation: item.representation || "",
					title: item.title || "",
					dataSet: item.dataSet ? JSON.parse(item.dataSet) : null,
					creation_date: item.creation_date
				};
			} catch (e) {
				log && log("_tmplReadInfo", "KO", item.id);
				return null;
			}
		},
		/**
		 * Resolve with [resources]/[]
		 */
		_tmplQuery: function(representation, endpoint, fail) {
			var self = this;
			return self._executeSql(_TMPL_QUERY)
				.then(function(result) {
					if (!result || !result.rows || result.rows.length == 0) return [];
					var resources = [],
						rec;
					for (var i = 0; i < result.rows.length; i++) {
						rec = self._tmplReadInfo(result.rows.item(i));
						if (rec) {
							resources.push(rec);
						}
					}
					log && log("_tmplQuery", "OK");
					return resources;
				}).fail(function(status) {
					_logError("_tmplQuery", status);
					return fail === true ? $.smReject(status) : [];
				});
		},
		/**
		 * Resolve with data/null
		 */
		_tmplRead: function(id) {
			var self = this;
			return self._executeSql(_TMPL_READ, [id])
				.then(function(result) {
					if (!result || !result.rows || result.rows.length == 0) return null;
					log && log("_tmplRead", id, "OK");
					return self._tmplReadInfo(result.rows.item(0));
				}).fail(function(status) {
					_logError("_tmplRead", status);
					return null;
				});
		},
		/**
		 * Resolve with  true/false
		 */
		_tmplDelete: function(id) {
			var self = this;
			return self._executeSql(_TMPL_DELETE, [id]).then(function(result) {
				log && log("_tmplDelete", id, result ? "OK" : "KO");
				return true;
			}).fail(function(status) {
				_logError("_tmplDelete", status);
				return false;
			});
		},
		/**
		 * Resolve with true/false
		 */
		_tmplDeleteAll: function(representation, endpoint) {
			var self = this;
			return self._executeSql(_TMPL_DELETE_ALL).then(function(result) {
				log && log("_tmplDeleteAll OK");
				return true;
			}).fail(function(status) {
				_logError("_tmplDeleteAll", status);
				return false;
			});
		},
		/**
		 * Resolve with template context or reject with an exception
		 */
		_tmplSave: function(ctx) {
			var self = this;
			var deferred = $.Deferred();
			if (!ctx || !ctx.id) {
				log && log("_tmplSave error", "Invalid parameters");
				deferred.reject(new Error("Invalid parameters"));
			} else {
				self._tmplDelete(ctx.id).then(function() {
					return self._executeSql(_TMPL_INSERT, [ctx.id, ctx.endpoint, ctx.representation, ctx.title, JSON.stringify(ctx.dataSet), ctx.creation_date]);
				}).then(function() {
					log && log("_tmplSave", ctx.id, "OK");
					deferred.resolve(ctx);
				}).fail(function(status) {
					_logError("_tmplSave", status);
					deferred.reject(status.$exception || new Error(status.message));
				});
			}
			return deferred.promise();
		},
		_tmplCount: function(endpoint, representation) {
			return this._executeSql(_TMPL_COUNT, []).then(function(result) {
				return result && result.rows.length > 0 ? result.rows.item(0).counter : 0;
			}).fail(function(status) {
				_logError("_tmplCount", status);
				return 0;
			});
		},
		/**
		 * Templates
		 * op: Operation
		 */
		templateOperation: function(op) {
			switch (op) {
				case "save":
					return this._tmplSave(arguments[1]);
				case "delete":
					return this._tmplDelete(arguments[1]);
				case "deleteAll":
					return this._tmplDeleteAll(arguments[1], arguments[2]);
				case "read":
					return this._tmplRead(arguments[1]);
				case "query":
					return this._tmplQuery(arguments[1], arguments[2], arguments[3]);
				case "count":
					return this._tmplCount(arguments[1], arguments[2]);
				default:
					return $.smReject(storageInterface.buildErrResult("Unknown template operation[" + op + "]"));
			}
		}
	});

exports.Klass = _Klass;
