// syracuse-ui/lib/desktop/site/desktopTopPanel.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Article = require("syracuse-ui/lib/common/article/article").Article;


var _localize = {
    $about: "About",
    $drafts: "Drafts",
    $historic: "History",
    $sessions: "Sessions",
    $bookmarks: "Bookmarks",
    $dashboards: "Dashboard",
    $bookmarkLinkTitle: "Add this page?",
    $bookmarkLinkDesc: "It will appear in your bookmark",
    $authoringLinkTitle: "Design the current page?",
    $authoringLinkDesc: "Launch the authoring panel"
};

function DesktopTopPanel(){

}

exports.DesktopTopPanel = helpers.defineClass(DesktopTopPanel, Article, {
    loadBox: function(){
        document.itemFactory.initializeItem(this, {});
        document.site.setZIndex(this.$$item = this.$$container);
        Article.prototype.loadBox.call(this);
        var $$item = document.site.userProfile ? document.site.userProfile.$$item : null;
        if ($$item) {
            if ($$item.parent().length == 0) {
                this.$$topUserProfile.empty().append($$item);
            }
        }
        this.$$container.css("height", (this.$$body.outerHeight() + this.$$body.offset().top) + "px");
    },
    drawBox: function(){
        this.$$body = $("<div id='s-site-top-pn-body'/>").appendTo(this.$$item);
        this.renderTopBar();
        this.renderNavigationBar();
        this.renderAboutBar();
    },
    renderTopLink: function(options){
        var dom = document.createElement("a");
        dom.className = "s-top-link";
        dom.setAttribute("href", "#");
        var $$link = $(dom);
        dom = document.createElement("div");
        dom.className = "s-top-link-icon " + options.$icon;
        $$link.append(dom);
        
        dom = document.createElement("div");
        dom.className = "s-top-link-title";
        $$link.append($(dom).text(options.$title).append($("<div class='s-top-link-desc'/>").text(options.$desc)));
        $$link.bind("click", options.$click).appendTo(this.$$topLinks);
    },
    renderTopBar: function(){
        var self = this;
        self.$$topBar = $("<div id='s-site-top-pn-top-bar'/>").appendTo(self.$$body);
        self.$$topUserProfile = $("<div id='s-site-top-pn-user-profile'/>").appendTo(self.$$topBar);
        
        self.$$topLinks = $("<div id='s-site-user-top-pn-links'/>").appendTo(self.$$topBar);
        self.renderTopLink({
            $icon: "s-top-link-bookmark-open",
            $title: _localize.$bookmarkLinkTitle,
            $desc: _localize.$bookmarkLinkDesc,
            $click: function(){
                return false;
            }
        });
        self.renderTopLink({
            $icon: "s-top-link-author-open",
            $title: _localize.$authoringLinkTitle,
            $desc: _localize.$authoringLinkDesc,
            $click: function(){
                document.site.toggleAuthoring(true);
                return false;
            }
        });
    },
    renderNavigationPanel: function(navItem){
        var div = document.createElement("div");
        div.className = "s-site-top-pn-nav-bar-cell " + navItem.$css;
        navItem.$$cell = $(div).appendTo(this.$$navigationBar);
        
        div = document.createElement("div");
        div.className = "s-site-top-pn-nav-title";
        navItem.$$title = $(div).text(navItem.$title).appendTo(navItem.$$cell);
        
        div = document.createElement("div");
        div.className = "s-site-top-pn-nav-item-body";
        navItem.links = document.itemFactory.load($(div).appendTo(navItem.$$cell), {
            $category: "links",
            $isBindDisabled: navItem.$isBindDisabled || false,
            $links: navItem.$links,
            $skin: "s-site-top-pn-nav-bar-menus"
        }, this);
        navItem.$$cell.bind("mouseleave mouseenter", function(event){
            navItem.links.$$item.find("a").toggleClass("s-site-top-pn-nav-bar-cell-over", event.type == "mouseenter");
            return false;
        });
    },
    renderNavigationBar: function(){
        var self = this;
        self.$$navigationBar = $("<div id='s-site-top-pn-nav-bar'/>").appendTo(self.$$body);
        Object.keys(self.navItems = {
            drafts: {
                $title: _localize.$drafts,
                $css: "s-site-top-pn-drafts"
            },
            historic: {
                $title: _localize.$historic,
                $isBindDisabled: true,
                $css: "s-site-top-pn-historic",
                $links: {
                    "banks": {
                        "$title": "List of banks"
                    },
                    "properties": {
                        "$title": "Bank properties (Boursorama)"
                    },
                    "newAsset": {
                        "$title": "New asset"
                    },
                    "listAsset": {
                        "$title": "List of assets"
                    },
                    "account": {
                        "$title": "Edit bank account (BNP)"
                    }
                }
            },
            bookmarks: {
                $title: _localize.$bookmarks,
                $css: "s-site-top-pn-bookmarks"
            },
            dashboards: {
                $title: _localize.$dashboards,
                $css: "s-site-top-pn-dashboards",
                $isBindDisabled: true,
                $links: self.$prototype.$navigation.$dashboard
            }
        }).forEach(function(id){
            self.renderNavigationPanel(self.navItems[id]);
        });
    },
    renderAboutBar: function(){
        this.$$aboutBar = $("<div id='s-site-top-pn-about-bar'/>").appendTo(this.$$body);
        document.itemFactory.load(this.$$aboutBar, {
            $category: "block",
            $title: _localize.$about,
            $skin: "s-site-top-pn-about-bar",
            $layout: {
                $items: [{
                
                    $category: "links",
                    $isBindDisabled: true,
                    $links: this.$prototype.$navigation.$about,
                    $skin: "s-site-top-pn-about-menus"
                }]
            }
        }, this);
    },
    
    dispose: function(){
        var self = this;
        var $$item = document.site.userProfile ? document.site.userProfile.$$item : null;
        if ($$item) {
            $$item.detach();
        }
        if (self.navItems) {
            Object.keys(self.navItems).forEach(function(prop){
                var navItem = self.navItems[prop];
                if (navItem && navItem.$$cell) {
                    navItem.$$cell.unbind();
                }
            });
        }
        Article.prototype.dispose.call(this);
    }
});
