// syracuse-ui/lib/desktop/site/dialog.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function Dialog(){
}

exports.Dialog = helpers.defineClass(Dialog, null, {
    open: function(options){
        if ((this.options = options).content) {
            if (this.options.position) {
                this._wrapPopup();
            }
            else {
                this._wrapDialogContent();
            }
        }
        else {
            this._openDialogPage();
        }
    },
    _wrapPopup: function(){
        var self = this;
        var domPopup = self.options.$$popup[0];
        document.site.setZIndex(self.options.$$popup);
        self._namespace = self.options.$$popup.attr("id") || document.controller.generateUUID();
        domPopup.style.position = "absolute";
        domPopup.setAttribute("id", self._namespace);
        self._boundarySelector = "#" + self._namespace;
        if (self.options.boundarySelector) {
            self._boundarySelector += "," + self.options.boundarySelector;
        }
        domPopup.style.display = '';
        if (self.options.position) {
            document.site.$$popupContainer.append(self.options.$$popup);
            self._reposition();
            document.site.addResizeListener(self);
        }
        $(document).bind("click." + self._namespace, function(event){
            var $$btn = $(event.target);
            if (self.options.$$popup.is(":visible")) {
                if ($$btn.closest(self._boundarySelector).length == 0) {
                    self.close();
                }
            }
            return true;
        });
    },
    _wrapDialogContent: function(){
        this.$$dialog = (this._content = this.options.content).$$item;
        this._ensureMode();
        this.$$dialog.appendTo(document.site.$$container);
        this.appendOverlay();
        document.site.setZIndex(this.$$dialog);
        this.onWindowResize();
        document.site.addResizeListener(this);
    },
    _openDialogPage: function(){
        var self = this;
        document.controller.loadRepresentation(self.options.article, self.options.$url, function($itemPage){
            self.$$dialog = $itemPage.$$container = $(document.createElement("div"));
            self._content = document.itemFactory.loadPage($itemPage);
            self._content.onAfterApplyChange = function(){
                self._reposition();
            };
            self._ensureMode();
            self.$$dialog[0].className = self.$skin + "-page-container";
            self._appendCloseButton();
            self._content.$$header[0].className = self.$skin + "-page-head";
            self._content.$$dataSlot[0].className = self.$skin + "-page-data";
            self._content.$$item[0].className = self.$skin + "-page";
            self._content.$$item.hide();
            self._content.$$body.css({
                overflow: "auto",
                height: 0
            });
            if (self.options.onValidate) {
                self._appendOkButton();
            }
            if (self.options.onSelectRecord) {
                self._content.onSelectRecord = function(selectedRecords){
                    if (self.options.onSelectRecord(selectedRecords, self._content) !== false) {
                        self.close();
                    }
                };
            }
            self.$$dialog.appendTo(document.site.$$container);
            self._content.$$item.show();
            self.appendOverlay();
            document.site.setZIndex(self.$$dialog);
            self.onWindowResize();
            document.site.addResizeListener(self);
        });
    },
    _ensureMode: function(){
        this.$skin = "s-" + (this._content.$dialogMode = this.options.$dialogMode || "modal");
    },
    appendOverlay: function(){
        if (this._content.$dialogMode == "modal") {
            var overlay = document.createElement("div");
            overlay.className = "s-overlay";
            document.site.setZIndex(this.$$overlay = $(overlay));
            this.$$overlay.appendTo(document.site.$$container);
        }
    },
    close: function(){
        if (this.options && this.options.closeHandler ? this.options.closeHandler() : true) {
            document.controller.disposeObject(this);
        }
    },
    _appendOkButton: function(){
        var self = this;
        var btn = document.createElement("a");
        btn.className = self.$skin + "-page-ok";
        self.$$okBtn = $(btn).bind("click", function(){
            if (self.options.onValidate(self._content) !== false) {
                self.close();
            }
            return false;
        }).appendTo(self._content.$$header);
    },
    _appendCloseButton: function(){
        var self = this;
        var btn = document.createElement("a");
        btn.className = self.$skin + "-page-close";
        self.$$closeBtn = $(btn).bind("click", function(){
            self.close();
            return false;
        }).appendTo(self._content.$$header);
    },
    onWindowResize: function(){
        if (this.$$overlay) {
            this.$$overlay.css({
                width: 0,
                height: 0
            }).css({
                width: document.site.$$container.outerWidth(),
                height: document.site.$$container.outerHeight()
            });
            this.$$overlay.show();
        }
        this._reposition();
    },
    _reposition: function(){
        if (this.options.position) {
            /* if (this.options.$size = "content") {
             var popup = this.options.$$popup[0];
             popup.style.width = popup.scrollWidth + "px";
             popup.style.height = popup.scrollHeight + "px";
             }*/
            this.options.$$popup.position(helpers.object.clone(this.options.position));
        }
        else {
            var body = document.site.$$body[0];
            var style = this.$$dialog[0].style;
            var left, top;
            var bodyWidth = document.site.$$container.innerWidth();
            var bodyHeight = document.site.$$container.innerHeight();
            switch (this.options.$dialogSize) {
                case "content":
                    left = (body.clientWidth - this.$$dialog.outerWidth()) / 2;
                    top = (body.clientHeight - this.$$dialog.outerHeight()) / 2;
                    break;
                default:
                    var margin = {
                        left: 30,
                        top: 30
                    };
                    left = margin.left;
                    top = document.site.$$body.offset().top + margin.top;
                    var width = body.clientWidth - (margin.left * 2);
                    if (width < 0) {
                        left = 0;
                        width = body.clientWidth;
                    }
                    style.width = width + "px";
                    var height = body.clientHeight - (margin.top * 2);
                    if (height < 0) {
                        top = 0;
                        height = body.clientHeight;
                    }
                    style.height = height + "px";
                    break;
            }
            style.left = left + "px";
            style.top = top + "px";
            if (this.options.$dialogSize != "content") {
                this._resizeDialogPage();
            }
        }
    },
    _resizeDialogPage: function(){
        var body = this._content.$$body[0];
        var root = this.$$dialog[0];
        var top = 0;
        var start = body;
        while (start && start != root) {
            top += start.offsetTop;
            start = start.offsetParent;
        }
        body.style.height = (this.$$dialog.height() - top) + "px";
    },
    dispose: function(){
        if (this.$$overlay) {
            this.$$overlay.remove();
        }
        document.site.removeResizeListener(this);
        if (this.$$closeBtn) {
            this.$$closeBtn.unbind();
            delete this.$$closeBtn;
        }
        if (this.$$okBtn) {
            this.$$okBtn.unbind();
            delete this.$$okBtn;
        }
        if (this.$$dialog) {
            this.$$dialog.empty().remove();
            delete this.$$dialog;
        }
        if (this.options) {
            if (this.options.position) {
                $(document).unbind("click." + this._namespace);
            }
            if (this.options.content) {
                //feree external content
                delete this.options.content;
            }
            else {
                //content is internal
                document.controller.disposeObject(this._content);
            }
            delete this.options.article;
            delete this.options.onValidate;
            delete this.options.onSelectRecord;
        }
        delete this._content;
    }
});
