// syracuse-ui/lib/desktop/site/dialogDecorator.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function DialogDecorator(){

}

exports.DialogDecorator = helpers.defineClass(DialogDecorator, null, {
    wrap: function(view){
        this.view = view;
        document.site.dialogAdapter = this
        this.$$overlay = $("<div class='s-overlay'/>").appendTo(document.site.$$container);
        if (view.$isItemDialog) {
            this.$$dialog = view.$$item;
        }
        else {
            this.$$dialog = $("<div/>").addClass(view.$cssDialog).append(view.$$item);
        }
        this.$$dialog.appendTo(document.site.$$container);
        this.$$overlay.css("z-index", parseInt(view.$$item.css("z-index") || "4000") - 10);
        this.onDocumentResize();
        return this;
    },
    onDocumentResize: function(){
        this.maximizeOverlay();
        this.$$overlay.show();
        this.center();
    },
    maximizeOverlay: function(){
        this.$$overlay.css({
            width: 0,
            height: 0
        }).css({
            width: document.site.$$container.outerWidth(),
            height: document.site.$$container.outerHeight()
        });
    },
    center: function(){
        var body = document.site.$$body[0];
        var style = this.$$dialog[0].style;
        var left, top;
        var bodyWidth = document.site.$$container.innerWidth();
        var bodyHeight = document.site.$$container.innerHeight();
        switch (this.view.$dialogSize) {
            case "content":
                left = (body.clientWidth - this.$$dialog.outerWidth()) / 2;
                top = (body.clientHeight - this.$$dialog.outerHeight()) / 2;
                break;
            default:
                debugger;
                var margin = {
                    left: 30,
                    top: 30
                };
                left = margin.left;
                top = document.site.$$body.offset().top + margin.top;
                var width = body.clientWidth - (margin.left * 2);
                if (width < 0) {
                    left = 0;
                    width = body.clientWidth;
                }
                style.width = width + "px";
                var height = body.clientHeight - (margin.top * 2);
                if (height < 0) {
                    top = 0;
                    height = body.clientHeight;
                }
                style.height = height + "px";
                break;
        }
        style.left = left + "px";
        style.top = top + "px";
        if (this.view.onDialogResize) {
            this.view.onDialogResize(this.$$dialog);
        }
    },
    dispose: function(){
        if (this.$$overlay) {
            this.$$overlay.remove();
        }
        if (this.$$dialog) {
            this.$$dialog.remove();
            delete this.$$dialog;
        }
        delete this.view;
    }
});
