// syracuse-ui/lib/desktop/site/messageBox.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

var _localize = {
    msgboxok: "Ok",
    msgboxyes: "Yes",
    msgboxno: "No",
    msgboxcancel: "Cancel"
};
function MessageBox(){
}

exports.MessageBox = helpers.defineClass(MessageBox, null, {
    open: function(options){
        var self = this;
        self.$$btns = {};
        var domItem = document.createElement("div");
        domItem.className = "s-msgbox";
        self.$$item = $(domItem);
        self.callback = options.callback;
        
        var div = document.createElement("div");
        div.className = "s-msgbox-header";
        var $$header = $(div).appendTo(self.$$item);
        
        div = document.createElement("div");
        div.className = "s-msgbox-header-title";
        $(div).appendTo($$header).text(options.$title)
        
        div = document.createElement("div");
        div.className = "s-msgbox-body s-msgbox-msg-" + options.$type;
        self.$$body = $(div).appendTo(self.$$item);
        self.$$body.html(options.$message.replace(/\n/g, "<br/>"));
        
        div = document.createElement("div");
        div.className = "s-msgbox-foot";
        self.$$footer = $(div).appendTo(self.$$item);
        
        if (options.$$container) {
            (self.$$container = options.$$container).append(self.$$item);
            domItem.style.display = "inline-block";
        }
        else {
            domItem.style.position = "fixed";
            self.dialog = document.site.openDialog({
                id: document.controller.generateUUID(),
                content: self,
                $dialogSize: "content"
            });
        }
        
        self.drawButtons(options);
        
        self.$$footer.delegate("a.s-msgbox-button", "click", function(){
            var response = {
                $$button: $(this),
                $id: $(this).attr("data-s-id")
            };
            if (self.$$prompt) {
                response.prompt = self.$$prompt.val();
            }
            self.close(response.$id, response);
            return false;
        });
        
        if (options.$autoClose) {
            var count = (typeof(options.$autoClose) == "number" ? options.$autoClose : 4000) / 1000;
            div = document.createElement("div");
            div.className = "s-msgbox-header-timer";
            var $$timer = $(div).prependTo($$header).text("( " + count + " )")
            self.autoCloseTimeOut = setInterval(function(){
                $$timer.text("( " + (--count) + " )")
                if (count == 0) {
                    clearInterval(self.autoCloseTimeOut);
                    self.close("auto");
                }
            }, 1000);
        }
        $(document).bind("keypress.msgbox", function(evt){
            switch (evt.keyCode) {
                case 13:
                    setTimeout(function(){
                        self.$$defaultBtn.click();
                    }, 100);
                    break;
                case 27:
                    setTimeout(function(){
                        self.close("esc");
                    }, 100);
                    break;
            }
        });
    },
    drawButtons: function(options){
        var self = this;
        var $buttons = options.$buttons;
        if (options.$type == "question") {
            $buttons = $buttons || "yesno";
        }
        if (options.$type == "prompt") {
            $buttons = $buttons || "okcancel";
            self.$$prompt = $("<input type='text'/>").addClass("s-msgbox-prompt").attr("size", "50").appendTo(self.$$body);
        }
        
        $buttons = $buttons || "ok";
        ["yes", "no", "ok", "cancel"].forEach(function($id){
            if ($buttons.indexOf($id) >= 0) {
                self.drawButton({
                    $id: $id,
                    $text: _localize["msgbox" + $id]
                });
            }
        });
        
        self.$$defaultBtn = self.$$btns[options.$default || "ok"];
        if (!self.$$defaultBtn) {
            self.$$defaultBtn = self.$$btns[Object.keys(self.$$btns)[0]];
        }
        self.$$defaultBtn.addClass("s-msgbox-button-default");
    },
    drawButton: function($btn){
        var btn = document.createElement("a");
        btn.className = "s-msgbox-button";
        btn.setAttribute("data-s-id", $btn.$id);
        this.$$btns[$btn.$id] = $(btn).text($btn.$text).appendTo(this.$$footer);
    },
    close: function(closedBy, response){
        if (this.callback ? (this.callback(response, closedBy) !== false) : true) {
            if (this.$$item) {
                this.$$item.remove();
            }
            document.controller.disposeObject(this);
        }
    },
    dispose: function(){
        if (this.autoCloseTimeOut) {
            clearInterval(this.autoCloseTimeOut);
            delete this.autoCloseTimeOut;
        }
        if (this.$$footer) {
            this.$$footer.undelegate();
        }
        if (this.dialog) {
            this.dialog.close();
            delete this.dialog;
        }
        $(document).unbind("keypress.msgbox");
    }
});
