// syracuse-ui/lib/desktop/site/popup.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function Popup(){
}

exports.Popup = helpers.defineClass(Popup, null, {
    load: function(options){
        var self = this;
        var domPopup = options.$$popup[0];
        if (options.page) {
            var $$view = options.page.$$item;
            while ($$view && $$view[0].tagName.toUpperCase() != "BODY") {
                var zIndex = $$view.css("z-index");
                if (zIndex && zIndex.toLowerCase() != "auto") {
                    domPopup.style.zIndex = parseInt(zIndex) + 100;
                    break;
                }
                $$view = $$view.parent();
            }
        }
        if (domPopup.style.zIndex == "") {
            domPopup.style.zIndex = 4000;
        }
        
        self.options = options;
        self._namespace = self.options.$$popup.attr("id") || document.controller.generateUUID();
        domPopup.style.position = "absolute";
        domPopup.setAttribute("id", self._namespace);
        self._boundarySelector = "#" + self._namespace;
        if (self.options.boundarySelector) {
            self._boundarySelector += "," + self.options.boundarySelector;
        }
        
        domPopup.style.display = '';
        if (self.options.position) {
            document.site.$$popupContainer.append(self.options.$$popup);
            /* if (options.$isResizable) {
             self.options.$$popup.resizable();
             }*/
            self.resize();
            $(window).bind("resize." + self._namespace, function(evt){
                self.resize();
            });
        }
        
        $(document).bind("click." + self._namespace, function(event){
            var $$btn = $(event.target);
            if (self.options.$$popup.is(":visible")) {
                if ($$btn.closest(self._boundarySelector).length == 0) {
                    self.close();
                }
            }
            return true;
        });
        return this;
    },
    resize: function(){
        /* if (this.options.$size = "content") {
         var popup = this.options.$$popup[0];
         popup.style.width = popup.scrollWidth + "px";
         popup.style.height = popup.scrollHeight + "px";
         }*/
        this.options.$$popup.position(helpers.object.clone(this.options.position));
    },
    close: function(){
        if (this.options && this.options.closeHandler ? this.options.closeHandler() : true) {
            document.controller.disposeObject(this);
        }
    },
    dispose: function(){
        $(window).unbind("resize." + this._namespace);
        $(document).unbind("click." + this._namespace);
        if (this.options && this.options.$$popup) {
            this.options.$$popup.detach();
            this.options.$$popup[0].style.display = 'none';
        }
        //document.site.$$popupContainer.empty();
        delete this.options;
    }
});
