// @sage/ui-sql-designer/public/base.js
'use strict';
let requestTypes = {
    ADDTABLE: 'AT',
    DELETETABLE: 'DT',
    //MOVETABLE: 'MT',
    //RESIZETABLE: 'RT',
    PLACETABLE: 'TP',
    DELETEALL: 'CQ',
    CREATELINK: 'CL',
    DELETELINK: 'DL',
    LINKTYPE: 'LT',
    ADDFIELD: 'AF',
    MOVEFIELD: 'MF',
    LEVELFIELD: 'LF',
    DELETEFIELD: 'DF',
    CHANGEFIELDPROPERTY: 'CF',
    ENTEREXPRESSION: 'EE',
    MODIFYEXPRESSION: 'ME',
    SAVECONFIG: 'SC'
};
let linkType = {
    LEFT_JOIN: 'L',
    INNER_JOIN: 'I'
}
let changeFieldProperty = {
    BOUNDS: 'Bounds',
    VISIBLE: 'Visible',
    SORT: 'Sort',
    SUM: 'Sum',
    GROUP: 'Group',
    LEVEL: 'Level'
}
let changefieldType = {
    MOVE: 'M',
    DELETE: 'D'
}
let createElement = function(elementType, parent, id, className, textContent) {
    let el = document.createElement(elementType);
    if (id) {
        el.setAttribute('id', id);
    }
    if (className) {
        el.className = className;
    }
    if (parent) {
        parent.appendChild(el);
    }
    if (textContent) {
        el.textContent = textContent;
    }
    return el;
}

exports.requestTypes = requestTypes;
exports.linkType = linkType;
exports.changeFieldProperty = changeFieldProperty;
exports.changefieldType = changefieldType;
exports.createElement = createElement;
