// syracuse-tablet/html/js/wc/controls/vignette/ctrlVignetteLink.js
"use strict";

var Base = require('syracuse-tablet/html/js/wc/controls/ctrlSdataBase').CtrlSdataBase;

var utils = require('syracuse-tablet/html/js/wc/helpers/utils');
var Gadget = require('syracuse-tablet/html/js/wc/controls/vignette/gadget');
var fontUtils = require('syracuse-tablet/html/js/wc/ui/fontUtils');
var navigationPageStack = require('syracuse-tablet/html/js/wc/app/navigationPageStack');

var _templates = {
	linkPage: '\
			<section>\
				<div draggable="false" href="{{href}}"\
				class="s-m-link-page" data-nevent="" data-naction="touch">\
					{{#if icon}}\
						<section class="s-m-icon">\
							<div>\
								<div class="{{icon}}"></div>\
							</div>\
						</section>\
					{{/if}}\
					<section class="s-m-title">\
						<div>{{title}}</div>\
					</section>\
				</div>\
			</section>'
};

var _getHtml = function(name, ctx) {
	var key = name + "compiled";
	var tmpl = _templates[key];
	if (!tmpl) {
		tmpl = _templates[key] = Handlebars.compile(_templates[name]);
	}
	return tmpl(ctx);
};

/**
 * 
 * The prototype of vignettes contains an $item node which includes all the gadgets information
 */
exports.CtrlVignetteLink = utils.defineClass(
	function CtrlVignetteLink(controller, article, prototype, options) {
		Base.call(this, controller, article, prototype, options);

		/* Parameter values passed to the dashboard so they can be used to filter the gadgets use for test only
		options.dashboardParameters = {
			"CODE": {
				value: "10"
			}
		};
    	*/
		this.gadget = Gadget.createGadget(this.prototype, options.dashboardParameters);
		if (!this.gadget) {
			// TODO: Throw breaks widget tree creation, would be better to render an "error link" here 
			throw new Error("Cant' create vignette's gadget[" + this.id + "] - Check prototype")
		}
	}, Base, {
		destroy: function() {
			Base.prototype.destroy.call(this);
			if (this.navigationPageStack) {
				this.navigationPageStack.destroy();
			}
			this.navigationPageStack = null;
			if (this.gadget) {
				this.gadget.destroy();
			}
			this.gadget = null;

		},
		/**
		 */
		buildHtml: function() {
			// s-m-vignette for css compatibility
			Base.prototype.buildHtml.call(this, ["s-m-vignette"]);
			switch (this.gadget.getType()) {
				case "$representation":
					this._loadRepresentation();
					break;
				case "$dashboard":
					this._loadDashboard();
					break;
				case "$application":
					this._loadApplication();
					break;
				case "$external":
					this._loadExternal();
					break;
				default:
					break;
			}
		},

		afterRender: function() {},

		_getTitle: function() {
			return this.article && (this.article.$title ? this.getArticleText("$title") : this.gadget.data("$title"));
		},

		_loadRepresentation: function() {
			this._addPicker({
				"title": this._getTitle(),
				"icon": this._getCssIcon("s-m-sage-gadget-icon")
			});
		},

		_loadDashboard: function() {
			this._addPicker({
				"title": this._getTitle(),
				"icon": this._getCssIcon("s-m-sage-gadget-icon")
			});
		},

		_loadApplication: function() {
			this._addPicker({
				"title": this._getTitle(),
				"icon": this._getCssIcon("s-m-sage-app-icon"),
			});
		},

		_loadExternal: function() {
			var url = this.gadget.getExternalUrl();
			this._addPicker({
				"title": this._getTitle()
			});
		},

		_addPicker: function(context) {
			var vignetteCtx = {};
			for (var p in context) {
				vignetteCtx[p] = context[p];
			}
			$(_getHtml("linkPage", vignetteCtx)).appendTo(this.$$elmt);
		},

		_getCssIcon: function(sageBaseCss) {
			var icon = this.article && this.article.$icon || this.gadget.data("icon")
			if (icon) {
				if (icon.smStartsWith("s-m-")) {
					// our own gadget icons
					icon = sageBaseCss + " " + icon;
				} else {
					icon = fontUtils.getIconByName(icon);
				}
			}
			return icon || "";
		},

		_actTouch: function() {
			var link = this.gadget && this.gadget.getLink();
			if (link) {
				if (this.gadget.getType() == "$external") {
					window.open(link.$url, "s-m-external-window");
				} else {
					// Navigate by asking the controller because controller will be
					// able to switch between working copy and stateless urls 
					this.controller.openLink(link);
				}
			}
		},
		isVignette: function() {
			// The control is ctrlVignetteLink but since it's basically rendering a link and there is no "sub" content,
			// it's not actually a vignette
			return false;
		}
	});
