// syracuse-ui/lib/fusion/field/documentField.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var DocumentField = require('syracuse-ui/lib/field/documentField').DocumentField;

function CvgDocumentField() {}

exports.CvgDocumentField = helpers.defineClass(CvgDocumentField, DocumentField, {
	render: function() {
		DocumentField.prototype.render.call(this);
		syra_menus.button.disable(this.removeBtn, true);
	},
	setValue: function(value, metaData) {
		if (!this.isImage && value && value.$url) {
			this.$dataUrl = this.$field.$url = value.$url;
		}
		DocumentField.prototype.setValue.apply(this, value, metaData);
	},
	_onEndUpload: function(file) {
		DocumentField.prototype._onEndUpload.apply(this, arguments);
		this.page.externalAdapter.onUploadDone({
			"field": this,
			"file": file
		});
	}
});


function FusionImageField() {}

exports.FusionImageField = helpers.defineClassEx(FusionImageField, DocumentField, {
	initialize: function() {
		DocumentField.prototype.initialize.call(this);
		this.acceptFilter = "image/jpg,image/jpeg,image/png,image/gif,image/bmp";
	},
	getValue: function() {
		if (this.page.isFusionPage && this.articleParent.$fusionPageMeta) {
			return this.articleParent.$fusionPageMeta.winId + this.$item.$bind;
		}
		return null;
	}
});
