// syracuse-ui/lib/fusion/field/htmlVignetteField.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var JSXml = require('jsxml/lib/jsxml');
var VignetteField = require('syracuse-ui/lib/page/landing/vignette/field/htmlVignetteField');

function HTMLField() {}
exports.HTMLVignetteField = helpers.defineClass(HTMLField, VignetteField.HTMLVignetteField, {
	setFocus: function() {
		return true;
	},
	isDirty: function(value, record, metaData, $bind) {
		return false;
	}
});


function DHTMLField() {}
exports.DHTMLVignetteField = helpers.defineClass(DHTMLField, HTMLField, {
	setDataBind: function(value, record, meta) {
		var jsonX, urlParams;
		try {
			if ((jsonX = JSXml.parse(value))) {
				// At time we just support first url
				if ((urlParams = jsonX.DESC && jsonX.DESC.URL)) {
					urlParams = this._getFullUrl(typeof urlParams == "object" ? urlParams : urlParams[0], jsonX.DESC.TAD);
					VignetteField.HTMLVignetteField.prototype.setDataBind.call(this, this._getFullUrl(urlParams.ADR, urlParams.TAD), record, meta);
				}
			}
		} catch (e) {}
	},
	_getFullUrl: function buildUrl(url, type) {
		switch (type) {
			case "ABS":
				return url;
			case "PUB":
				//return this.getPubLocation() + "/" + url;
			case "DOS":
				//return this.getDosLocation() + "/" + url;
			case "LOC":
				return "";
			default:
				return url;
		}
	}
});
