// syracuse-ui/lib/fusion/field/htmlVignetteField.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var JSXml = require('jsxml/lib/jsxml');
var VignetteField = require('syracuse-ui/lib/field/vignette/htmlVignetteField');


function HTMLField() {}
exports.HTMLVignetteField = helpers.defineClass(HTMLField, VignetteField.HTMLVignetteField, {
	setFocus: function() {
		return true;
	},
	isDirty: function() {
		return false;
	}
});


function DHTMLField() {}
exports.DHTMLVignetteField = helpers.defineClass(DHTMLField, HTMLField, {
	setValue: function(value, meta) {
		var jsonX, urlParams;
		try {
			if ((jsonX = JSXml.parse(value))) {
				// At time we just support first url
				if ((urlParams = jsonX.DESC && jsonX.DESC.URL)) {
					urlParams = Array.isArray(urlParams) ? urlParams[0] : urlParams;
					VignetteField.HTMLVignetteField.prototype.setValue.call(this, this._getFullUrl(urlParams.ADR, urlParams.TAD), meta);
				}
			}
		} catch (e) {}
	},
	_getFullUrl: function buildUrl(url, type) {
		var fusionSess, $fusionController;
		switch (type) {
			case "ABS":
				return url;
			case "PUB":
			case "DOS":
				$fusionController = syra_fusion.syraUtil.getFusionController(this),
				fusionSess = $fusionController ? $fusionController.getSession() : null;
				return fusionSess ? fusionSess.getAppServerHttpUrl(type == "PUB", url) : "about:blank";
			case "LOC":
				return "";
			default:
				return url;
		}
	}
});
