// syracuse-tablet/html/js/controls/panels/footer/panelFooterEditHomeSmartphone.js
"use strict";

var Base = require('syracuse-tablet/html/js/controls/panels/footer/panelFooterSdataPageBase').PanelFooterSdataPageBase;
var utils = require('syracuse-tablet/html/js/helpers/utils');
var actionAdapter = require('syracuse-tablet/html/js/pages/actionAdapter');
var panelsHtml = require('syracuse-tablet/html/js/controls/panels/panelsHtml');

/**
 * Dedicated header for edit homePage on Tablet
 */
exports.PanelFooterEditHomeSmartphone = utils.defineClass(
	function PanelFooterEditHomeSmartphone(controller, type, article, opts) {
		Base.call(this, controller, type, article, opts);
	}, Base, {

		destroy: function() {
			Base.prototype.destroy.call(this);
		},
		buildHtml: function() {
			var self = this;

			// Base class creates footer and a left and right container that
			// are available after by this.$$left and this.$$right
			Base.prototype.buildHtml.call(self, "smartphone");

			this.leftItems = [];
			this.rightItems = [];

			var items = this._getItems();
			for (var ii = 0, jj = items.length; ii < jj; ii++) {
				var item = items[ii];
				var html = panelsHtml.getItemHtml("footer", item);
				if (html) {
					// Only store items here to be able to change order before adding to the dom
					var list = item.slot === "left" ? this.leftItems : this.rightItems;
					list.push(html);
				}
			}
			this.rightItems.reverse();
			this.$$left.append($(this.leftItems.join("")));
			this.$$right.append($(this.rightItems.join("")));
		},
		_getItems: function() {
			var items = [];
			items.push(
				actionAdapter.createDataAction("addGroup", "addGroup", null, "welcome.dashboard.add.group", "addGroup", this.controller, {
					slot: "left"
				})
			);
			if (this.actionAdapter.showMyAvailableDashboard) {
				items.push(
					actionAdapter.createDataAction("myAvailableTemplates", "myAvailableTemplates", null, "welcome.dashboard.myAvailableTemplates.list", "myAvailableTemplates", this.controller, {
						slot: "left"
					})
				);
			}
			return items;
		}
	});
