// syracuse-tablet/html/js/controls/panels/footer/panelFooterSdataPageTablet.js
"use strict";

var Base = require('syracuse-tablet/html/js/controls/panels/footer/panelFooterSdataPageBase').PanelFooterSdataPageBase;

var utils = require('syracuse-tablet/html/js/helpers/utils');
var panelsHtml = require('syracuse-tablet/html/js/controls/panels/panelsHtml');

/**
 * 
 */
exports.PanelFooterSdataPageTablet = utils.defineClass(
	function PanelFooterSdataPageTablet(controller, type, article, opts) {
		Base.call(this, controller, type, article, opts);
	}, Base, {

		destroy: function() {
			Base.prototype.destroy.call(this);
		},
		buildHtml: function() {
			var self = this;

			// Base class creates footer and a left and right container that
			// are available after by this.$$left and this.$$right
			Base.prototype.buildHtml.call(self, "tablet");

			this.leftItems = [];
			this.rightItems = [];

			var items = this._getItems();

			for (var ii = 0, jj = items.length; ii < jj; ii++) {
				var item = items[ii];
				if (this.controller.isActionHidden && this.controller.isActionHidden(item.action)) {
					continue;
				}
				var html = panelsHtml.getItemHtml("footer", item);
				if (html) {
					// Only store items here to be able to change order before adding to the dom
					var list = item.slot === "left" ? this.leftItems : this.rightItems;
					list.push(html);
				}
			}
			this.rightItems.reverse();
			this.$$left.append($(this.leftItems.join("")));
			this.$$right.append($(this.rightItems.join("")));
		},
		/**
		 * Can be overridden
		 */
		_getItems: function() {
			return this.actionAdapter.getFooterItems();
		}
	});
