// syracuse-ui/lib/tests/menus/menuTreeSettings.js
"use strict";
exports.$prototype = {
	"$properties": {
		$fieldJsonDelta: {
			"$title": "delta",
			"$type": "application/x-string"
		}
	},
	$links: {
		"pattern-1": {
			"$title": "tabs & rows"
		},
		"pattern-2": {
			"$title": "3 rows"
		},
		"pattern-3": {
			"$title": "4 rows"
		},
		"pattern-4": {
			"$title": "1 column - 4 rows"
		},
		"bo-1": {
			$url: "http://www.sage.fr",
			"$title": "Bienvenue",
			"$description": "Retxqs dcsq dQFqdfDQ fdsqfq  sqdfsdqf sq sdqfsdqfsdq"
		},
		"bo-2": {
			"$title": "Icon 1",
			"$description": "Retxqs dcsq dQFqdfDQ fdsqfq  sqdfsdqf sq sdqfsdqfsdq"
		},
		"bo-3": {
			"$title": "Icon 2",
			"$description": "Retxqs dcsq dQFqdfDQ fdsqfq  sqdfsdqf sq sdqfsdqfsdq"
		},
		"bo-4": {
			"$title": "Icon 3",
			"$description": "Retxqs dcsq dQFqdfDQ fdsqfq  sqdfsdqf sq sdqfsdqfsdq"
		},
		"bo-5": {
			"$title": "Icon 4",
			"$description": "Retxqs dcsq dQFqdfDQ fdsqfq  sqdfsdqf sq sdqfsdqfsdq"
		},
		"bo-6": {
			"$title": "Icon 5",
			"$description": "Retxqs dcsq dQFqdfDQ fdsqfq  sqdfsdqf sq sdqfsdqfsdq"
		},
		"sage-1": {
			"$title": "Les enjeux de la centralisation de trésorerie",
			"$description": "Retxqs dcsq dQFqdfDQ fdsqfq  sqdfsdqf sq sdqfsdqfsdq"
		},
		"sage-2": {
			"$title": "Sage et la mobilité (en anglais).",
			"$description": "Sage vous présente une infographie vidéo sur la mobilité dans le monde (en aglais)"
		},
		"sage-31": {
			"$title": "RSE et Marque Employeur : l'attractivité RH en ligne de mire",
			"$description": "RSE & la marque employeur : L'attractivité RH en ligne de mire Témoignage de Célia Vedovotto, consultante Sénior chez ConvictionRH",
			"$url": "http://www.youtube.com//watch?v=57G917DBEPQ&list=UUQgiNxZN-Gl2CIw1Djg3qDg&index=4"
		},
		"sage-4": {
			"$title": "SAGE SOLUTIONERS CROISSANCE",
			"$url": "http://www.youtube.com/watch?v=QuzEwlECrgM&list=UUQgiNxZN-Gl2CIw1Djg3qDg&index=5"
		},
		"sage-5": {
			"$title": "SAGE SOLUTIONERS GESTION",
			"$url": "http://www.youtube.com/watch?v=wG6TdP2Out0&list=UUQgiNxZN-Gl2CIw1Djg3qDg&index=6"
		},
		"sage-6": {
			"$title": "SAGE SOLUTIONERS RENTABILITE",
			"$url": "http://www.youtube.com/watch?v=d24DOql6dm0&list=UUQgiNxZN-Gl2CIw1Djg3qDg&index=7"
		},
		"sage-7": {
			"$title": "Sage vous souhaite une bonne année 2013",
			"$description": "Sage vous souhaite une bonne et heureuse année 2013.",
			"$url": "http://www.youtube.com/watch?v=ENKWJUDrD-c&list=UUQgiNxZN-Gl2CIw1Djg3qDg&index=8"
		},
		"sage-8": {
			"$title": "Lancement d'un projet ERP",
			"$description": "Isabelle Saint-martin, chef de marché ERP chez Sage division moyennes et grandes entreprises, répond aux questions de DFCG TV lors du Financium 2012. L'interview porte sur ...",
			"$url": "http://www.youtube.com/watch?v=XDcGGzWHRzM&list=UUQgiNxZN-Gl2CIw1Djg3qDg&index=9"
		},
		"$addSection": {
			"$title": "$addSection"
		},
		"$addBlock": {
			"$title": "$addBlock"
		},
		"$fieldLinkfirstName": {
			"$title": "First Name"
		},
		"$fieldLinklastName": {
			"$title": "Last Name"
		},
		"$fieldLinktitle": {
			"$title": "Title"
		},
		"$fieldLinkmaritalStatus": {
			"$title": "Marital Status"
		},
		"$fieldLinkbirthDate": {
			"$title": "Birth Date"
		},
		"$fieldLinktestTime": {
			"$title": "Test Time"
		},
		"$fieldLinkforeignerWorker": {
			"$title": "Foreigner worker"
		},
		"$fieldLinkemail": {
			"$title": "Email"
		},
		"$fieldLinkphoneNumber": {
			"$title": "Phone Number"
		},
		"$fieldLinkphoto": {
			"$title": "Photo"
		},
		"$fieldLinkaddressStreet": {
			"$title": "Street"
		},
		"$fieldLinkaddressZipCode": {
			"$title": "Zip code"
		},
		"$fieldLinkaddressCity": {
			"$title": "City"
		},
		"$fieldLinkcomment": {
			"$title": "Comment"
		},
		"$fieldLinktestInteger": {
			"$title": "Test Integer"
		},
		"$fieldLinktestDecimal": {
			"$title": "Test Decimal"
		},
		"$fieldLinkcountry": {
			"$title": "Country"
		},
		"$fieldLinkcompany": {
			"$title": "Company"
		},
		"$fieldLinkchildren": {
			"$title": "Children"
		},
		"$fieldLink$creUser": {
			"$title": "Created by"
		},
		"$fieldLink$updUser": {
			"$title": "Updated by"
		},
		"$fieldLink$creDate": {
			"$title": "Created on"
		},
		"$fieldLink$updDate": {
			"$title": "Updated on"
		}

	},
	$actions: {
		menutest2: {
			$title: "$isSelected record 2",
			testcase: {
				$layout: {
					$items: [{
						"$category": "menus",
						$isHidden: false,
						"$title": "selectTemplate",
						"$skin": "s-mn-bar-extra-icon",
						"$isBoxCollapsable": "true",
						"$isPopupContent": true,
						"$titleIcon": {
							"$value": "template",
							"$path": "authoring/s-layout-"
						},
						"$itemIcon": {
							"$mode": "iconText",
							"$path": "authoring/s-layout-"
						},
						"$layout": {
							"$items": [{
								"$bind": "pattern-1"
							}, {
								"$bind": "pattern-2"
							}, {
								"$bind": "pattern-3"
							}, {
								"$bind": "pattern-4"
							}]
						}
					}]
				}
			}
		},
		menutest5: {
			$title: "Sage, les vidéos mises en ligne",
			testcase: {
				"$category": "menus",
				"$skin": "s-test-sage-menus",
				$title: "Sage, les vidéos mises en ligne",
				"$isBoxCollapsable": "true",
				$isDescriptionVisible: true,
				"$itemIcon": {
					"$mode": "iconText",
					"$path": "test/"
				},
				$layout: {
					$items: [{
						$bind: "sage-1"
					}, {
						$bind: "sage-2"
					}, {
						$bind: "sage-3"
					}, {
						$bind: "sage-4"
					}, {
						$bind: "sage-5"
					}, {
						$bind: "sage-6"
					}, {
						$bind: "sage-7"
					}, {
						$bind: "sage-8"
					}]
				}
			}
		}
	}
};

exports.$layout = {
	$layoutType: "row",
	$widths: "70,30",
	$items: [{
		"$category": "section",
		$title: "Result",
		$clientId: "$result",
		"$layout": {
			"$items": []
		}
	}, {
		$layoutType: "stack",
		$items: [{
			"$layoutType": "tabs",
			"$items": [{
				"$category": "section",
				$title: "Select test case",
				"$layout": {
					"$items": [{
						"$category": "menus",
						"$clientId": "$samples",
						"$skin": "s-mn-bar-extra",
						"$itemSkin": "s-mn-bar-extra-link",
						"$layout": {
							$items: [{
								$bind: "menutest2"
							}, {
								$bind: "menutest3"
							}, {
								$bind: "menutest5"
							}]
						}
					}]
				}
			}, {
				"$category": "section",
				"$title": "Json Delta",
				"$layout": {
					"$items": [{
						"$bind": "$fieldJsonDelta",
						"$isTitleHidden": true,
						"$isEditMode": true,
						"$rows": 30,
						"$skin": "s-json-field"
					}]
				}
			}]
		}]

	}]
};
