// syracuse-ui/lib/tests/menus/menuTreeSettings.js
exports.$prototype = {
    "$properties": {
        $fieldJsonDelta: {
            "$title": "delta",
            "$type": "application/x-string"
        }
    },
    $actions: {
        menutest2: {
            $title: "$isSelected record 2",
            testcase: {
                $category: "links",
                $id: "formatLayoutLinks",
                $skin: "s-aw-designer-links",
                $isBindDisabled: true,
                $choiceLayout: "1",
                $links: {
                    "composite-3-1": {
                        "$title": "tabs & rows"
                    },
                    "composite-3-2": {
                        "$title": "3 rows"
                    },
                    "composite-4-1": {
                        "$title": "4 rows"
                    },
                    "composite-1-3": {
                        "$title": "1 column - 4 rows"
                    }
                },
                $menus: [{
                    $items: [{
                        $title: "selectTemplate",
                        $icon: {
                            $mode: "icon",
                            $value: "template",
                            $path: "authoring/s-layout-",
                            $width: "25px",
                            $height: "25px"
                        },
                        $subMenus: {
                            $skin: "s-aw-composites",
                            $isPopup: true,
                            $icon: {
                                $mode: "iconText",
                                $path: "authoring/s-layout-",
                                $width: "25px",
                                $height: "25px"
                            }
                        },
                        $items: [{
                            "$bind": "composite-3-1"
                        }, {
                            "$bind": "composite-3-2"
                        }, {
                            "$bind": "composite-4-1"
                        }, {
                            "$bind": "composite-1-3"
                        }]
                    }]
                }]
            }
        },
        menutest1: {
            $title: "upadate title array",
            testcase: {
                "$category": "menus",
                "$id": "$samples",
                "$skin": "s-test-menus",
                $isRoot: true,
                "$layout": {
                    $items: [{
                        $bind: "menutest1"
                    }, {
                        $bind: "menutest2"
                    }]
                }
            }
        }
    }
};

exports.$layout = {
    $layoutType: "row",
    $items: [{
        "$category": "block",
        $title: "Result",
        $id: "$result",
        "$layout": {
            "$items": []
        }
    }, {
        "$layoutType": "tabs",
        "$items": [{
            "$category": "block",
            $title: "Select test case",
            "$layout": {
                "$items": [{
                    "$category": "menus",
                    "$id": "$samples",
                    "$skin": "s-test-menus",
                    $isRoot: true,
                    "$layout": {
                        $items: [{
                            $bind: "menutest1"
                        }, {
                            $bind: "menutest2"
                        }]
                    }
                }]
            }
        }, {
            "$category": "section",
            "$title": "Json Delta",
            "$layout": {
                "$items": [{
                    "$bind": "$fieldJsonDelta",
                    "$isTitleHidden": true,
                    "$isEditMode": true,
                    "$rows": 30,
                    "$skin": "s-json-field"
                }]
            }
        }]
    }]
};
