// syracuse-ui/lib/tests/menus/menusTest.js
"use strict";
var helpers = require("syracuse-core/lib/helpers");
var TestPage = require("syracuse-ui/lib/tests/testPage").TestPage;

function MenusTest() {}

exports.MenusTest = helpers.defineClass(MenusTest, TestPage, {
	defineMenu: function($title, $skin) {
		var $items = Object.keys(this.$prototype.$links).map(function($bind) {
			return {
				$bind: $bind
			};
		});
		Object.keys(this.$prototype.$actions).forEach(function($bind) {
			$items.push({
				$bind: $bind
			});
		});
		this.$item.$layout.$items.push({
			$category: "section",
			$title: $title,
			$layout: {
				$items: [{
					$category: "menus",
					$skin: $skin,
					$layout: {
						$items: $items
					}
				}]
			}
		});
	},
	_initiliazeTestPage: function() {
		var self = this;
		["$links", "$actions", "$linksParameters", "$actionsParameters"].forEach(function($value) {
			var $menus = self.getResources().getActionLinks($value);
			["$links", "$actions"].forEach(function($type) {
				if ($menus[$type]) {
					Object.keys($menus[$type]).forEach(function($bind) {
						self.$prototype[$type][$bind] = $menus[$type][$bind];
					});
				}
			});
		});
		this.$item.$layout = {
			$items: []
		};
		this.defineMenu("s-mn-main", "s-mn-main");
		this.defineMenu("s-mn-tools", "s-mn-tools");
		this.defineMenu("s-mn-extra", "s-mn-extra");
		this.defineMenu("s-mn-business", "s-mn-business");
		this.defineMenu("s-mn-prefer", "s-mn-prefer");
		this.defineMenu("s-vignette-menus", "s-vignette-menus");
		this.defineMenu("s-vignette-widget-menus", "s-vignette-widget-menus");
		this.defineMenu("s-list-menus (popup)", "s-list-menus");
		this.defineMenu("s-list-menus-record-menus (popup)", "s-list-menus-record-menus");
		this.defineMenu("s-field-menus (popup)", "s-field-menus");
		this.defineMenu("s-site-header-menus", "s-site-header-menus");

	}
});
