// syracuse-ui/lib/tests/menus/menuTreeSettings.js
"use strict";
exports.$prototype = {
	"$properties": {
		$fieldJsonDelta: {
			"$title": "delta",
			"$type": "application/x-string"
		}
	},
	$links: {
		"pattern-1": {
			"$title": "tabs & rows"
		},
		"pattern-2": {
			"$title": "3 rows"
		},
		"pattern-3": {
			"$title": "4 rows"
		},
		"pattern-4": {
			"$title": "1 column - 4 rows"
		},
		"bo-1": {
			$url: "http://www.services.bouyguestelecom.fr/telephone-mobile/bienvenue-chez-bouygues-telecom",
			"$title": "Bienvenue chez Bouygues Telecom",
			"$description": "Retrouvez en un clic toutes les applications Bouygues Telecom réunies"
		},
		"bo-2": {
			"$title": "Mon Espace Synchro",
			"$description": "Sauvegardez gratuitement vos contacts et contenus multimédia"
		},
		"bo-3": {
			"$title": "Application Sécurité",
			"$description": "Protégez votre mobile des applications malveillantes"
		},
		"bo-4": {
			"$title": "Contrôle parental",
			"$description": "Activez le contrôle parental mobile pour encadrer le surf de vos enfants"
		},
		"bo-5": {
			"$title": "Anti-Spam mobile",
			"$description": "Protégez votre mobile contre les spams SMS ou vocaux"
		},
		"bo-6": {
			"$title": "Ici Info",
			"$description": "Retrouvez en un coup d'oeil tout ce qu'il y a autour de vous : restos, cinés, métros!!!"
		},
		"sage-1": {
			"$title": "Les enjeux de la centralisation de trésorerie",
			"$description": "Le contexte SEPA favorise au sein des entreprises la mise en place de projets de centralisation de trésorerie. Une démarche qui pose des enjeux d'ordre opérationnels et tec...",
			"$url": "http://www.youtube.com/watch?v=Z3c_Fhzlpw8&list=UUQgiNxZN-Gl2CIw1Djg3qDg&index=1"
		},
		"sage-2": {
			"$title": "Sage et la mobilité (en anglais).",
			"$description": "Sage vous présente une infographie vidéo sur la mobilité dans le monde (en aglais).",
			"$url": "http://www.youtube.com/watch?v=6gawGBHJsSw&list=UUQgiNxZN-Gl2CIw1Djg3qDg&index=2"
		},
		"sage-31": {
			"$title": "RSE et Marque Employeur : l'attractivité RH en ligne de mire",
			"$description": "RSE & la marque employeur : L'attractivité RH en ligne de mire Témoignage de Célia Vedovotto, consultante Sénior chez ConvictionRH",
			"$url": "http://www.youtube.com//watch?v=57G917DBEPQ&list=UUQgiNxZN-Gl2CIw1Djg3qDg&index=4"
		},
		"sage-4": {
			"$title": "SAGE SOLUTIONERS CROISSANCE",
			"$url": "http://www.youtube.com/watch?v=QuzEwlECrgM&list=UUQgiNxZN-Gl2CIw1Djg3qDg&index=5"
		},
		"sage-5": {
			"$title": "SAGE SOLUTIONERS GESTION",
			"$url": "http://www.youtube.com/watch?v=wG6TdP2Out0&list=UUQgiNxZN-Gl2CIw1Djg3qDg&index=6"
		},
		"sage-6": {
			"$title": "SAGE SOLUTIONERS RENTABILITE",
			"$url": "http://www.youtube.com/watch?v=d24DOql6dm0&list=UUQgiNxZN-Gl2CIw1Djg3qDg&index=7"
		},
		"sage-7": {
			"$title": "Sage vous souhaite une bonne année 2013",
			"$description": "Sage vous souhaite une bonne et heureuse année 2013.",
			"$url": "http://www.youtube.com/watch?v=ENKWJUDrD-c&list=UUQgiNxZN-Gl2CIw1Djg3qDg&index=8"
		},
		"sage-8": {
			"$title": "Lancement d'un projet ERP",
			"$description": "Isabelle Saint-martin, chef de marché ERP chez Sage division moyennes et grandes entreprises, répond aux questions de DFCG TV lors du Financium 2012. L'interview porte sur ...",
			"$url": "http://www.youtube.com/watch?v=XDcGGzWHRzM&list=UUQgiNxZN-Gl2CIw1Djg3qDg&index=9"
		},
		"$addSection": {
			"$title": "$addSection"
		},
		"$addBlock": {
			"$title": "$addBlock"
		},
		"$fieldLinkfirstName": {
			"$title": "First Name"
		},
		"$fieldLinklastName": {
			"$title": "Last Name"
		},
		"$fieldLinktitle": {
			"$title": "Title"
		},
		"$fieldLinkmaritalStatus": {
			"$title": "Marital Status"
		},
		"$fieldLinkbirthDate": {
			"$title": "Birth Date"
		},
		"$fieldLinktestTime": {
			"$title": "Test Time"
		},
		"$fieldLinkforeignerWorker": {
			"$title": "Foreigner worker"
		},
		"$fieldLinkemail": {
			"$title": "Email"
		},
		"$fieldLinkphoneNumber": {
			"$title": "Phone Number"
		},
		"$fieldLinkphoto": {
			"$title": "Photo"
		},
		"$fieldLinkaddressStreet": {
			"$title": "Street"
		},
		"$fieldLinkaddressZipCode": {
			"$title": "Zip code"
		},
		"$fieldLinkaddressCity": {
			"$title": "City"
		},
		"$fieldLinkcomment": {
			"$title": "Comment"
		},
		"$fieldLinktestInteger": {
			"$title": "Test Integer"
		},
		"$fieldLinktestDecimal": {
			"$title": "Test Decimal"
		},
		"$fieldLinkcountry": {
			"$title": "Country"
		},
		"$fieldLinkcompany": {
			"$title": "Company"
		},
		"$fieldLinkchildren": {
			"$title": "Children"
		},
		"$fieldLink$creUser": {
			"$title": "Created by"
		},
		"$fieldLink$updUser": {
			"$title": "Updated by"
		},
		"$fieldLink$creDate": {
			"$title": "Created on"
		},
		"$fieldLink$updDate": {
			"$title": "Updated on"
		}

	},
	$actions: {
		menutest2: {
			$title: "$isSelected record 2",
			testcase: {
				$layout: {
					$items: [{
						"$category": "menus",
						$isHidden: false,
						"$title": "selectTemplate",
						"$skin": "s-test-menus-icon",
						"$isBoxCollapsable": "true",
						"$isPopupContent": true,
						"$titleIcon": {
							"$value": "template",
							"$path": "authoring/s-layout-"
						},
						"$itemIcon": {
							"$mode": "iconText",
							"$path": "authoring/s-layout-"
						},
						"$layout": {
							"$items": [{
								"$bind": "pattern-1"
							}, {
								"$bind": "pattern-2"
							}, {
								"$bind": "pattern-3"
							}, {
								"$bind": "pattern-4"
							}]
						}
					}]
				}
			}
		},
		menutest3: {
			$title: "bouygues",
			testcase: {
				"$category": "menus",
				$title: "bouygues",
				"$skin": "s-test-bo",
				$isDescriptionVisible: true,
				"$itemIcon": {
					"$mode": "iconText",
					"$path": "test/"
				},
				"$layout": {
					$layoutType: "row",
					$items: [{
						$layoutType: "stack",
						$items: [{
							$bind: "bo-1"
						}, {
							$bind: "bo-2"
						}]
					}, {
						$layoutType: "stack",
						$items: [{
							$bind: "bo-3"
						}, {
							$bind: "bo-4"
						}]
					}, {
						$layoutType: "stack",
						$items: [{
							$bind: "bo-5"
						}, {
							$bind: "bo-6"
						}]
					}]
				}
			}
		},
		menutest4: {
			$title: "columns",
			testcase: {
				"$category": "menus",
				"$skin": "s-test-menus",
				$layout: {
					$items: [{
						$bind: "$isSelected"
					}, {
						$layoutType: "row",
						$items: [{
							$title: "update Title",
							$layout: {
								$items: [{
									"$skin": "s-test-bo-link",
									$isDescriptionVisible: true,
									"$icon": {
										"$mode": "iconText",
										"$path": "test/"
									},
									$bind: "bo-5"
								}, {
									$bind: "testlink-childField"
								}, {
									"$skin": "s-test-bo-link",
									$isDescriptionVisible: true,
									"$icon": {
										"$mode": "iconText",
										"$path": "test/"
									},
									$bind: "bo-2"
								}, {
									"$skin": "s-test-menus-icon-link",
									"$icon": {
										"$mode": "iconText",
										"$path": "authoring/s-layout-"
									},
									$bind: "pattern-3"
								}]
							}
						}, {
							$title: "show /hide",
							$layout: {
								$items: [{
									"$skin": "s-test-menus-icon-link",
									"$icon": {
										"$mode": "iconText",
										"$path": "authoring/s-layout-"
									},
									$bind: "pattern-2"
								}, {
									$bind: "testlink-array"
								}, {
									"$skin": "s-test-menus-icon-link",
									"$icon": {
										"$mode": "iconText",
										"$path": "authoring/s-layout-"
									},
									$bind: "pattern-1"
								}, {
									$bind: "testlink-graph"
								}]
							}
						}]
					}, {
						$layoutType: "row",
						$items: [{
							$title: "Messages",
							$layout: {
								$items: [{
									$bind: "testlink-messagebox"
								}, {
									$bind: "testlink-messageBoxPopup"
								}, {
									$bind: "testlink-diagnoses"
								}]
							}
						}, {
							$title: "Menus",
							$skin: "s-mn-bar-main",
							"$isBoxCollapsable": "true",
							$layout: {
								$items: [{
									$bind: "testlink-menustree"
								}, {
									$bind: "testlink-menus"
								}, {
									$bind: "testlink-param"
								}]
							}
						}]
					}, {
						$title: "styles",
						$layout: {
							$items: [{
								$bind: "$fieldStylecolumn"
							}, {
								$bind: "$fieldStylerecords"
							}, {
								$bind: "$valueStyle"
							}]
						}
					}, {
						$layoutType: "row",
						$items: [{
							$title: "CRUD",
							$layout: {
								$items: [{
									$bind: "insert"
								}, {
									$bind: "updaterecord"
								}, {
									$bind: "updaterecords"
								}, {
									$bind: "reorder"
								}, {
									$bind: "deleterecord"
								}, {
									$bind: "deleterecords"
								}]
							}
						}, {
							$title: "Convergence",
							$layout: {
								$items: [{
									$bind: "testx3LinkAQFB"
								}, {
									$bind: "testx3LinkAQFB"
								}]
							}
						}]
					}, {
						$title: "Json",
						$layout: {
							$items: [{
								$bind: "JsonLint"
							}, {
								$bind: "jsonformatter"
							}]
						}
					}, {
						$title: "capabilities",
						$layout: {
							$items: [{
								$bind: "capabilities"
							}, {
								$bind: "capabilitiesinsertdel"
							}, {
								$bind: "capabilitiesnone"
							}]
						}
					}]
				}
			}
		},
		menutest5: {
			$title: "Sage, les vidéos mises en ligne",
			testcase: {
				"$category": "menus",
				"$skin": "s-test-sage-menus",
				$title: "Sage, les vidéos mises en ligne",
				"$isBoxCollapsable": "true",
				$isDescriptionVisible: true,
				"$itemIcon": {
					"$mode": "iconText",
					"$path": "test/"
				},
				$layout: {
					$items: [{
						$bind: "sage-1"
					}, {
						$bind: "sage-2"
					}, {
						$bind: "sage-3"
					}, {
						$bind: "sage-4"
					}, {
						$bind: "sage-5"
					}, {
						$bind: "sage-6"
					}, {
						$bind: "sage-7"
					}, {
						$bind: "sage-8"
					}]
				}
			}
		}
	}
};

exports.$layout = {
	$layoutType: "row",
	$widths: "70,30",
	$items: [{
		"$category": "section",
		$title: "Result",
		$clientId: "$result",
		"$layout": {
			"$items": []
		}
	}, {
		$layoutType: "stack",
		$items: [{
			"$layoutType": "tabs",
			"$items": [{
				"$category": "section",
				$title: "Select test case",
				"$layout": {
					"$items": [{
						"$category": "menus",
						"$clientId": "$samples",
						"$skin": "s-test-menus",
						"$layout": {
							$items: [{
								$bind: "menutest2"
							}, {
								$bind: "menutest3"
							}, {
								$bind: "menutest4"
							}, {
								$bind: "menutest5"
							}]
						}
					}]
				}
			}, {
				"$category": "section",
				"$title": "Json Delta",
				"$layout": {
					"$items": [{
						"$bind": "$fieldJsonDelta",
						"$isTitleHidden": true,
						"$isEditMode": true,
						"$rows": 30,
						"$skin": "s-json-field"
					}]
				}
			}]
		}]

	}]
};
