// @sage/syracuse-react/src/profile/panel/tabs/localesTabController.tsx
"use strict";
import * as React from 'react';
import { Ajax } from '../../../ajax/ajax';
import { LocalesTab } from './localesTab';
import { IProfile, IProfileLocaleSelect } from '../../bar/profileBarProps';
import { ILocalePreferences, } from '../../../culture/culture';
import * as Locales from '../../../culture/locales';
import * as SyraPageModal from '../../../over/modal/page/syraPageModal';
import { showDiagnoses } from '../../../alert/alert';
export interface ILocalesTabProps extends IProfile, IProfileLocaleSelect { }

export class LocalesTabController extends React.Component<ILocalesTabProps, {}> {
    private ajax: Ajax;
    constructor(props: any, context: any) {
        super(props, context);
        if (!props.profile.locales) {
            this.ajax = Locales.load((locales: ILocalePreferences[]) => {
                this.props.profile.locales = locales;
                this.forceUpdate();
            });
        }
    }
    onMoreClick = () => {
        SyraPageModal.openSelectModal({
            authoringDisabled: true,
            url: this.props.profile.localeLookupUrl,
            onLoadRepresentation: (diagnoses) => diagnoses && showDiagnoses(diagnoses),
            onValidateSelection: (selected: any[]) => this.props.onSelectLocale(selected[0])
        });
    }
    componentWillUnmount() {
        this.ajax && this.ajax.dispose();
    }
    render() {
        if (this.props.profile.locales) {
            return <LocalesTab locales={this.props.profile.locales}
                selected={this.props.profile.selectedLocale}
                onSelectLocale={this.props.onSelectLocale}
                onMoreClick={this.onMoreClick} />;
        }
        return <div>{"In loading"}</div>;
    }
}
