// @sage/syracuse-react/src/profile/panel/tabs/themesTab.tsx
"use strict";
import * as React from 'react';
import  {getLocalString} from '../../../culture/localString';
import { ScrollList } from '../../../common/list/list';
import { Link} from '../../../link/link';
import { ITheme, DEFAULT_THEME_UUID } from '../../../style/styleSheet';

export class ThemesTab extends React.Component<{
	selected: ITheme,
	themes: ITheme[],
	onSelectTheme: (theme: ITheme) => void
}, {
	}> {
	render() {
		let selectedId = this.props.selected && this.props.selected.$uuid;
		let unSelected: any[] = this.props.themes.filter((theme) => selectedId != theme.$uuid);
		if (selectedId) {
			//if default, selectedId == null
			unSelected.unshift({
				$uuid: DEFAULT_THEME_UUID,
				description: getLocalString("profile_theme_default")
			});
		}
		return <div className="s_profile_panel_tabs_body s_profile_themes">
			<ScrollList className="s_profile_panel_list">
				{unSelected.map((theme: ITheme, index: number) =>
					<Link
						className="s_profile_panel_list_item"
						key={index}
						title={theme.description}
						value={theme}
						onClick={this.props.onSelectTheme} />
				)}
			</ScrollList>
		</div>;
	}
}
