// syracuse-tablet/html/js/controls/page/panels/actionPanel.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var Base = require('syracuse-tablet/html/js/controls/page/panels/sidePanel').Klass;
var eventListener = require('syracuse-tablet/html/js/application/eventListener');
var notifications = require('syracuse-tablet/html/js/helpers/notifications');
var prototype = require('syracuse-tablet/html/js/helpers/prototype');
var HammerScroller = require('syracuse-tablet/html/js/controls/hammerScroller').Klass;
var uiRect = require('syracuse-tablet/html/js/ui/rect');
var globals = require('syracuse-tablet/html/js/helpers/globals');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var ActionManager = require('syracuse-tablet/html/js/application/actionManager').Klass;

var _templates = {
	main: '\
			<nav class="navbar s-m-side-panel {{side}} navbar-inverse ctrl-evt-click" role="navigation">\
				<div class="s-m-scroll-elmt">\
				{{#if facetlink}}\
					<div class="container-fluid">\
						<div class="navbar-header">\
							<span class="navbar-brand" href="#"><span>{{linksHeaderTitle}}</span></a>\
						</div>\
					</div>\
					<ul class="nav navbar-nav">\
						{{{facetLinksHtml}}}\
					</ul>\
				{{/if}}\
				<div class="container-fluid">\
					<div class="navbar-header">\
						<span class="navbar-brand" href="#"><span>{{title}}</span></a>\
					</div>\
				</div>\
					<ul class="nav navbar-nav">\
						{{{actionsHtml}}}\
					</ul>\
				</div>\
			</nav>',
	facetLinks: '\
		{{#each links}}\
			<li>\
		 		{{#if $isAction}} \
					<a draggable="false" href="#" class="{{css}}" data-action="actionLink" data-params="{{$uuid}}" data-control-id="{{../../ctrlId}}">\
						<i style="font-size: large" class="{{icon}}"/>\
						<span>{{title}}</span>\
					</a>\
				{{else}} \
					<a draggable="false" href="#" class="{{css}}" data-nav="{{page}}" data-params="{{$uuid}}" data-sdata-parameters="{{parameters}}" data-control-id="{{../ctrlId}}">\
						<i style="font-size: large" class="{{icon}}"/>\
						<span>{{title}}</span>\
					</a>\
				{{/if}} \
			</li>\
		{{/each}}'
};
/**
 * Base left/right panel control for dashBoards and regular pages
 * 	Adapted from Metro Bootstrap navbar-side
 */
var _Klass = utils.defineClass(

	function(controller, $type, side, options) {
		this.enableGesture = options && options.enableGesture;
		Base.call(this, controller, $type, side, options);
	}, Base, {

		destroy: function() {
			Base.prototype.destroy.call(this);
			if (this._gestureMgr) {
				this._gestureMgr.destroy();
				this._gestureMgr = null;
			}
		},

		/**
		 * Build the html here when we know the orientation and deviceClass
		 */
		onMainPageResize: function(pageInfo, orientation, deviceClass) {
			var context = {
				ctrlId: this.id,
				side: this.side
			};
			if (this.$$panel) {
				this.$$panel.remove();
			}
			if (this._gestureMgr) {
				this._gestureMgr.destroy();
				this._gestureMgr = null;
			}
			this.setPanel(this.buildAfterMainResize(context, pageInfo, orientation, deviceClass));
		},
		buildAfterMainResize: function(context, pageInfo, orientation, deviceClass) {
			// append facet links if any
			var deviceType = globals.getSiteLayout().getDeviceType(deviceClass);
			var facetLinks = this.controller.prototype.getLinks(this.getDao(), "actionpanelparent", {
				deviceType: deviceType
			});
			if (facetLinks && Object.keys(facetLinks).length > 0) {
				var actMgr = new ActionManager({
					$links: facetLinks,
					noMessage: false
				});
				context.facetlink = true;
				context.linksHeaderTitle = locale.text("actionpanel.label.links");
				context.facetLinksHtml = Handlebars.compile(_templates.facetLinks)({
					links: actMgr.getLinks()
				});
			}
			return $(Handlebars.compile(_templates.main)(context));
		},

		/**
		 * Toggle the panel
		 * 	action: toggle(click button footer), close(click dismiss) or resize (click resize icon inside the panel)
		 * 	callBack: shift the content for left panel
		 * 	This action is called by footer action and also by toggle data-action
		 */
		toggle: function(action, callBack) {
			Base.prototype.toggle.call(this, action, callBack);
			if (this.enableGesture) {
				this.updateGesture();
			}
		},
		updateGesture: function() {
			if (this.$$elmt.is(':visible')) {
				if (!this._gestureMgr) {
					this.$$scrollElmt = this.$$panel.find(".s-m-scroll-elmt");
					this.$$scrollElmt.css({
						height: "auto",
						position: "relative"
					});
					this.$$scrollWrapper = this.$$panel;
					this._gestureMgr = new HammerScroller(this.$$scrollElmt, {
						direction: "v",
						valMax: 0,
						name: this.$bind,
						isPageScroller: false
					});
				}
				var scrollRect = uiRect.elmtRect(this.$$scrollElmt, "outer");
				var wrapperRect = uiRect.elmtRect(this.$$scrollWrapper, "outer");
				var viewRect = wrapperRect.intersectRect(scrollRect);
				if (viewRect && !viewRect.contains(scrollRect)) {
					this._gestureMgr.init(viewRect);
				} else {
					this._gestureMgr.reset();
				}
			} else if (this._gestureMgr) {
				this._gestureMgr.reset();
			}
		}
	});

exports.Klass = _Klass;
