// syracuse-tablet/html/js/controls/page/panels/pageTemplatePanel.js
"use strict";

var globals = require('syracuse-tablet/html/js/helpers/globals');
var utils = require('syracuse-tablet/html/js/helpers/utils');
var notifications = require('syracuse-tablet/html/js/helpers/notifications');
var Base = require('syracuse-tablet/html/js/controls/page/panels/dialogPanel').Klass;
var locale = require('syracuse-tablet/html/js/helpers/locale');
var fontUtils = require('syracuse-tablet/html/js/ui/fontUtils');
var formatApi = require('syracuse-tablet/html/js/helpers/formatApi');

var _txt = function(id) {
	return locale.text("edit.templates.dlog." + id);
};
var _templates = {
	error: '<section class="draft-error"><h1>{{error}}<h1><h2>{{contact}}<h2></section>',
	main: '\
		<article style="height:100%">\
			<section class="panel panel-default create">\
				<div class="panel-heading">{{createSection}}</div>\
				<div class="panel-body">\
					<table style="width:100%"><tr>\
						<td>{{name}}</td>\
						<td>\
							<input class="ctrl-evt-input" type="text" style="width: 100%;"/>\
						</td>\
						<td>\
							<span style="float:right;" class="btn btn-default disabled" data-action="create" data-control-id="{{ctrlId}}">\
								<i class="{{createicn}}"></i>\
								<span>{{create}}</span>\
							</span>\
						</td>\
					</tr></table>\
				</div>\
			</section>\
			<section class="panel panel-default avail" style="overflow:hidden">\
				<div class="panel-heading"  style="width:100%">{{availSection}}</div>\
				<div style="overflow:hidden">\
					<ul style="margin:-1px" class="list-group s-m-scroll-elmt">\
					</ul>\
				</div>\
			</section>\
		</article>',
	items: '\
		{{#each items}}\
			<li class="list-group-item" data-params="{{tmplId}}">\
				<div>\
					<span style="display:inline-block;"><div><b>{{title}}</b></div><div><small>{{date}}</small></div></span>\
					<span style="float:right;">\
						<span class="btn btn-default " data-action="apply" data-control-id="{{../ctrlId}}">\
							<i class="{{../applyicn}}"></i>\
							<span>{{../apply}}</span>\
						</span>\
						<span class="btn btn-default" data-action="update" data-control-id="{{../ctrlId}}">\
							<i class="{{../updateicn}}"></i>\
						</span>\
						<span class="btn btn-default" data-action="delete" data-control-id="{{../ctrlId}}">\
							<i class="{{../deleteicn}}"></i>\
						</span>\
					</span>\
				</div>\
			</li>\
		{{/each}}'
};
var _getHtml = function(name, ctx) {
	var key = name + "compiled";
	var tmpl = _templates[key];
	if (!tmpl) {
		tmpl = _templates[key] = Handlebars.compile(_templates[name]);
	}
	if (!tmpl) alert("template " + name + " not found");
	return tmpl(ctx);
};

var _Klass = utils.defineClass(

	function(controller, $type, panelId, options) {
		options.enableGesture = true;
		Base.call(this, controller, $type, panelId, options);
		this.representation = controller.representation;
		this.dateFormatter = formatApi.getFormatter("application/x-datetime");
	}, Base, {
		buildHtml: function(side) {
			if (this.$$content) {
				this.$$content.remove();
			}
			Base.prototype.buildHtml.call(this, side, _txt("title"));
			this.$$content = this._buildContent().appendTo(this.$$elmt);
			this.$$list = this.$$content.find(".panel.avail ul");
		},
		/**
		 * Build the main content synchronously (without tmpl list)
		 * We need to build at least s-m-scroll-elmt synchronously to be compliant with sidePanel gesture management
		 */
		_buildContent: function(representation, endpoint) {
			var context = {
				createSection: _txt("createSection"),
				availSection: _txt("availSection"),
				name: _txt("name"),
				create: _txt("create"),
				createicn: fontUtils.getIconByName("$apply"),
				ctrlId: this.id
			};
			return $(_getHtml("main", context));
		},
		/**
		 * Build the list of templates asynchronously
		 */
		_buildItems: function(representation, endpoint) {
			var self = this;
			self.$$list.off().empty();
			var context = {
				ctrlId: self.id,
				updateicn: fontUtils.getIconByName("$refresh"),
				deleteicn: fontUtils.getIconByName("$delete"),
				applyicn: fontUtils.getIconByName("$apply"),
				apply: _txt("apply"),
				items: []
			};
			globals.getStorage().templateOperation("query", self.representation, globals.getEndpoint(), true).then(function(templates) {
				(templates || []).forEach(function(tmpl, idx) {
					context.items.push({
						title: tmpl.title,
						date: self.dateFormatter.formatValue(tmpl["creation_date"]),
						tmplId: tmpl.id
					});
				});
				$(_getHtml("items", context)).appendTo(self.$$list);
				// An update is needed because height changed
				self._updateGesture();
			}).fail(function(e) {
				globals.getModal().error(locale.text("edit.templates.dlog.readallerr"), e, function() {
					self.$$content.html($(_getHtml("error", {
						error: locale.text("error"),
						contact: locale.text("contact.admin")
					})));
				});
			});
		},
		refresh: function() {
			this._buildItems();
			this.$$content.find('input[type="text"]').val("");
		},
		_notify: function(text) {
			var notify = {
				severityClass: "success",
				body: text,
				onlyWebapp: true
			};
			globals.getModal().notify(notify);
		},
		_getTmplId: function($target, event) {
			event.stopPropagation();
			return $target.closest("li[data-params]").attr("data-params");
		},
		_actApply: function(param, $target, event) {
			var self = this;
			var tmplId = self._getTmplId($target, event);
			if (!tmplId) return;
			globals.getStorage().templateOperation("read", tmplId).then(function(data) {
				self._notify("Template has been read successfully");
				self.close();
				if (data) {
					self.controller.dao.setTemplateData(data.dataSet);
					self.controller.refresh({
						refreshValue: true
					});
				}
			}).fail(function(e) {
				globals.getModal().error("Error deleting template", e);
			});
		},
		_actUpdate: function(param, $target, event) {
			var self = this;
			var tmplId = self._getTmplId($target, event);
			if (!tmplId) return;
			globals.getStorage().templateOperation("update", tmplId, self.controller.dao.getTemplateData()).then(function() {
				self.close();
				self._notify("Template has been updated successfully");
			}).fail(function(e) {
				globals.getModal().error("Error updating template", e);
			});
		},
		_actDelete: function(param, $target, event) {
			var self = this;
			var tmplId = self._getTmplId($target, event);
			if (!tmplId) return;
			globals.getStorage().templateOperation("delete", tmplId).then(function() {
				self._notify("Template has been deleted successfully");
				self.refresh();
			}).fail(function(e) {
				globals.getModal().error("Error deleting template", e);
			});
		},
		_actCreate: function(param, $target, event) {
			var self = this;
			var ctx = {
				id: utils.UUID(),
				endpoint: globals.getEndpoint(),
				representation: self.representation,
				title: self.$$content.find('.panel.create input').val(),
				dataSet: self.controller.dao.getTemplateData()
			};
			globals.getStorage().templateOperation("save", ctx).then(function() {
				self.close();
				self._notify("Template has been created successfully");
			}).fail(function(e) {
				globals.getModal().error("Error saving template", e);
			});
		},
		/**
		 * Manages create button status
		 */
		onInput: function(evt) {
			var $$t = $(evt.target);
			if (!$$t.is("input")) return;
			if (!this._$$create) {
				this._$$create = this.$$content.find('.btn[data-action="create"]');
			}
			this._$$create.toggleClass("disabled", $$t.val().trim().length <= 0);
		},
		notifyToggled: function(side, action, visible) {
			if (!visible) return;
			this.refresh();
		}
	});

exports.Klass = _Klass;
