// syracuse-tablet/html/js/controls/page/panels/regularActionPanel.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var Base = require('syracuse-tablet/html/js/controls/page/panels/actionPanel').Klass;
var ActionManager = require('syracuse-tablet/html/js/application/actionManager').Klass;

var _templates = {
	actionLink: '\
		{{#each links}}\
			<li>\
		 		{{#if $isAction}} \
					<a draggable="false" href="#" class="{{css}}" data-action="actionLink" data-params="{{$uuid}}" data-control-id="{{../../ctrlId}}">\
						<i style="font-size: large" class="{{icon}}"/>\
						<span>{{title}}</span>\
					</a>\
				{{else}} \
					<a draggable="false" href="#" class="{{css}}" data-nav="{{page}}" data-params="{{$uuid}}" data-sdata-parameters="{{parameters}}" data-control-id="{{../ctrlId}}">\
						<i style="font-size: large" class="{{icon}}"/>\
						<span>{{title}}</span>\
					</a>\
				{{/if}} \
			</li>\
		{{/each}}'
};

var _getHtml = function(name, ctx) {
	var key = name + "compiled";
	var tmpl = _templates[key];
	if (!tmpl) {
		tmpl = _templates[key] = Handlebars.compile(_templates[name]);
	}
	return tmpl(ctx);
};

/**
 * Regular page action panel
 */
var _Klass = utils.defineClass(

	function(controller, $type, panelId, options) {
		Base.call(this, controller, $type, panelId, options);
	}, Base, {

		buildPanel: function(context, side) {
			if (this.actionManager) {
				// When we change the deviceClass we can have a non null actionManager
				this.actionManager.destroy();
			}
			this.actionManager = new ActionManager({
				$links: this.controller.prototype.getLinks(this.getDao(), "actionpanel"),
				noMessage: false
			});
			context.title = "Actions";
			context.actionsHtml = _getHtml("actionLink", {
				ctrlId: this.id,
				links: this.actionManager.getLinks()
			});
			return Base.prototype.buildPanel.call(this, context);
		},
		_actActionLink: function(uuid) {
			this.close("close");
			this.actionManager.triggerLink(uuid, this.controller).then(function(success, link, data, options) {}).fail(function() {});
		}
	});

exports.Klass = _Klass;
