// syracuse-tablet/html/js/controls/page/panels/userPanelSmartphone.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var Base = require('syracuse-tablet/html/js/controls/page/panels/actionPanel').Klass;
var globals = require('syracuse-tablet/html/js/helpers/globals');
var fontUtils = require('syracuse-tablet/html/js/ui/fontUtils');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var eventListener = require('syracuse-tablet/html/js/application/eventListener');
var notifications = require('syracuse-tablet/html/js/helpers/notifications');

var _templates = {
	navigLink: '<li class="">\
		<a draggable="false"  href="#" data-action="{{action}}" data-params="{{params}}" data-control-id="{{ctrlId}}">\
			<i class="{{icon}}"></i>\
			&nbsp;\
			<span>{{title}}</span>\
		</a>\
	</li>'
};

// TODO
var _panelLinks = {
	home: {
		label: "header.sm.links.home",
		icon: "home",
		action: "gotoWelcomeApplication",
		target: "application"
	},
	switchContext: {
		label: "header.sm.links.switchContext",
		icon: "switchContext",
		params: "switchContext",
		action: "linkClicked",
		target: "panel"
	},
	pageShare: {
		label: "header.links.page.share",
		icon: "nativeShare",
		params: "pageShare",
		action: "linkClicked",
		target: "panel"
	},
	clearCache: {
		label: "header.sm.links.settings",
		icon: "settings",
		action: "linkClicked",
		params: "settings",
		target: "panel"
	},
	about: {
		label: "header.sm.links.about",
		icon: "about",
		action: "linkClicked",
		params: "about",
		target: "panel"
	},
	logout: {
		label: "header.sm.links.logout",
		icon: "logout",
		action: "logout",
		target: "application"
	},
	development: {
		label: "header.links.development",
		icon: "developmentPage",
		action: "developmentPage",
		target: "application"
	},
	pinPage: {
		label: "",
		icon: "",
		action: "pinPage",
		target: "application"
	},
	pageConfig: {
		label: "page.configure.btn",
		icon: "pageConfig",
		action: "pageConfig",
		target: "application"
	},
	drafts: {
		label: "header.links.page.drafts",
		icon: "drafts",
		action: "pageDrafts",
		target: "application"
	}
};

var _getHtml = function(name, ctx) {
	var key = name + "compiled";
	var tmpl = _templates[key];
	if (!tmpl) {
		tmpl = _templates[key] = Handlebars.compile(_templates[name]);
	}
	return tmpl(ctx);
};

/**
 * Regular page left panel
 */
var _Klass = utils.defineClass(

	function(controller, $type, panelId, options) {
		options = options || {};
		// Just to test this mode that pushs the content
		options.overlap = false;
		Base.call(this, controller, $type, panelId, options);
		notifications.subscribe(this, ["sm.page.pin.state.change", "sm.drafts.changed"]);
		this._initPanelItems(controller, options);
	}, Base, {

		buildPanel: function(context, side) {
			var self = this;
			var actionsHtml = [];
			var keys = Object.keys(self._panelLinks);
			keys.forEach(function(kk, idx) {
				var link = self._panelLinks[kk];
				if (link.hidden !== true) {
					actionsHtml.push(_getHtml("navigLink", {
						title: locale.text(link.label),
						action: link.action || "",
						params: link.params || "",
						icon: fontUtils.pageIcon(link.icon),
						ctrlId: link.target === "panel" ? self.id : null
					}));
				}
			});
			context.items = [{
				title: globals.getUserName(),
				html: actionsHtml.join('')
			}];
			return Base.prototype.buildPanel.call(self, context, side);
		},

		_initPanelItems: function(page, options) {
			var toolBars = options || {};
			this._panelLinks = $.extend(true, {}, _panelLinks);
			this._panelLinks["development"].hidden = !toolBars.showDevelopment;
			this._panelLinks["home"].hidden = !toolBars.showHome;
			this._panelLinks["pageConfig"].hidden = !toolBars.showPageConfig;
			this._panelLinks["pageShare"].hidden = !page.showPageShare();
			//page.nativeCapabilities.share,
			var pinPage = this._panelLinks["pinPage"];
			switch (toolBars.statusPinPage) {
				case 1:
					pinPage.label = "pin.page.panel.label.pin";
					pinPage.hidden = false;
					pinPage.icon = "pinPage_1";
					break;
				case 2:
					pinPage.label = "pin.page.panel.label.unpin";
					pinPage.hidden = false;
					pinPage.icon = "pinPage_2";
					break;
				default:
					pinPage.hidden = true;
					break;
			}
		},

		_actLinkClicked: function(actionName) {
			/* triggering action */
			eventListener.triggerAction(this.$$elmt, actionName);
		},

		// newState
		// 0 = don't show
		// 1 = show inactive
		// 2 = show active
		notifPagePinStateChange: function(newState) {
			var $$icon = this.$$elmt.find("[data-action=pinPage] i");
			if ($$icon.length > 0) {
				$$icon.attr("class", fontUtils.pageIcon("pinPage_" + newState));
			}
			var $$text = this.$$elmt.find("[data-action=pinPage] span");
			if ($$text.length > 0) {
				var text;
				if (newState != 2) {
					text = locale.text("pin.page.panel.label.pin");
				} else {
					text = locale.text("pin.page.panel.label.unpin");
				}

				$$text.text(text);
			}
		},
		hasLinks: function() {
			return true;
		},
		notifDraftsChanged: function(count) {
			if (!this.$$elmt) return;
			var $$e = this.$$elmt.find("[data-params=pageShare]");
			var $b = $$e.find("span.badge");
			if ($b.length === 0) {
				$b = $('<span class="badge"/>').appendTo($$e);
			}
			$b.text(count);
		}
	});

exports.Klass = _Klass;
