// @sage/syracuse-react/src/office/officeAddin.ts
"use strict";
import { getLocalString } from '../culture/localString';
import { alertBox } from '../alert/alert';
import * as syra_expression from '../syra/utility/syra_expression';
import * as Cookie from '../utility/storage/cookie';

export const exe = "/msoffice/lib/general/addIn/SyracuseOfficeAddinsSetup.EXE";

export function installed(menu: any, article: any, title: any, end: () => void) {
	if (menu.$confirm && menu.$officeAddinSetup) {
		if (Cookie.get("SyracuseOfficeAddin2") === "valid") {
			return true;
		}
	}
	if (menu.$url && menu.$url.indexOf("SyracuseOfficeAddinsSetup.EXE") > 0) {
		let date = new Date();
		Cookie.set("SyracuseOfficeAddin2", "valid", null, 3650);
	}
	return menu.$confirm ? confirm(menu, article, title, end) : true;
};

function confirm(menu: any, article: any, title: any, end: () => void) {
	window.setTimeout(() => {
		alertBox({
			severity: "info",
			autoClose: menu.$officeAddinSetup ? 10000 : 0,
			title: syra_expression.parse(article, title || getLocalString("siteDefaultConfirmTitle")),
			message: syra_expression.parse(article, menu.$officeAddinSetup ? getLocalString("installOfficeAddin") : menu.$confirm),
			buttons: {
				mode: "yes,no"
			},
			onClose: (closeBy: string) => {
				if (closeBy == "yes") {
					if (menu.$officeAddinSetup) {
						let date = new Date();
						Cookie.set("SyracuseOfficeAddin2", "valid", null, 3650);
					}
					end();
				}
			}
		});
	}, 50);
	return false;
}
