// syracuse-tablet/html/js/wc/controls/array/builderHeader.js
"use strict";

var utils = require('syracuse-tablet/html/js/wc/helpers/utils');
var FiltersSort = require('syracuse-tablet/html/js/wc/controls/array/filtersSort');

var _html = {
	header: '<header> \
			<nav class="s-m-stdheader"> \
				{{#if showPagination}} \
				<nav class="s-m-pagin"> \
					<ul class="pagination"> \
						<li> \
							<a class="s-m-link glyphicon glyphicon-step-backward disabled" href="#" \
								data-nevent \
								data-naction="pagination" \
								data-params="$first" \
								draggable="false"> \
							</a> \
						</li> \
						<li> \
							<a class="s-m-link glyphicon glyphicon-chevron-left disabled" href="#" \
								data-nevent \
								data-naction="pagination" \
								data-params="$previous" \
								draggable="false"> \
							</a> \
						</li> \
						<li> \
							<a class="s-m-link glyphicon glyphicon-chevron-right disabled" href="#" \
								data-nevent \
								data-naction="pagination" \
								data-params="$next" \
								draggable="false "> \
							</a> \
						</li> \
						<li> \
							<a class="s-m-link glyphicon glyphicon-step-forward disabled" href="#" \
								data-nevent \
								data-naction="pagination" \
								data-params="$last" \
								draggable="false"> \
							</a> \
						</li> \
					</ul> \
				</nav> \
				{{/if}} \
				{{#if showSearch}} \
				<nav class="s-m-search"/>\
				{{/if}} \
				{{#if showSortFilter}} \
				<div class="s-m-filterSort" data-nevent data-naction="actionArray" data-params="panelfiltersort">\
					<i class="s-m-sorted" style="display:none"></i>\
					<i class="s-m-filtered fa fa-filter" style="display:none"></i>\
					<i class="fa fa-bars"></i>\
				</div>\
				{{/if}}\
				{{#if showFiltersList}}\
				<nav class="s-m-filterslist"/>\
				{{/if}}\
			</nav> \
			{{#if showFiltersTabs}}\
				<nav class="s-m-filtertabs"/>\
			{{/if}}\
		</header>'
};

var _templates = {};
var _getHtml = function(name, ctx) {
	var tmpl = _templates[name];
	if (!tmpl) {
		tmpl = _templates[name] = Handlebars.compile(_html[name]);
	}
	return tmpl(ctx || {});
};

/**
 * Render array header
 */
exports.BuilderHeader = utils.defineClass(
	function builderHeader(control) {
		this.control = control;
	}, null, {

		destroy: function() {
			this.control = null;
			if (this.$$elmt) {
				this.$$elmt.remove()
			}
			this.$$elmt = null;
		},

		buildHtml: function() {
			if (this.$$elmt) {
				this.$$elmt.remove();
			}
			var context = $.extend(true, null, this.control.article.$arrayOptions);
			var $filtAuth = this.control._filters && this.control._filters.$filtersAllowed() ? this.control.$filtersGetAuthoring() : "none";
			context.showFiltersList = $filtAuth === "list";
			context.showFiltersTabs = $filtAuth === "tabs";
			context.showSearch = context.showSearch && this.control._search != null;
			context.showSortFilter = context.showSortFilter && FiltersSort.filterSortAllowed(this.control);

			if (context.showFiltersList || context.showFiltersTabs || context.showSearch || context.showSortFilter || context.showPagination) {
				this.$$elmt = $(_getHtml("header", context));
				if (context.showSearch) {
					$("nav.s-m-search", this.$$elmt).append(this.control._search.getHtml())
				}
				if (context.showFiltersList) {
					$("nav.s-m-filterslist", this.$$elmt).append(this.control._filters.getHtml("list"));
				} else if (context.showFiltersTabs) {
					$("nav.s-m-filtertabs", this.$$elmt).append(this.control._filters.getHtml("tabs"));
				}
			} else {
				// Empty
				this.$$elmt = $("<header/>");
			}

			this.control.$$content.before(this.$$elmt);
			this.update();
			// Check if header empty

			return this.$$elmt;
		},

		update: function() {
			if (this.control.paginator) {
				var links = this.control.paginator.pagingLinks;
				var self = this;
				links.forEach(function(link) {
					var $$link = $('[data-naction="pagination"][data-params="' + link + '"]', self.$$elmt);
					var ena = self.control.paginator.isEnabled(link);
					$$link.toggleClass("disabled", !ena);
				});
			}
		}

	});
