// syracuse-tablet/html/js/wc/controls/array/builderChart.js
"use strict";

var utils = require('syracuse-tablet/html/js/wc/helpers/utils');
var globals = require('syracuse-tablet/html/js/wc/app/globals');
var locale = require('syracuse-tablet/html/js/wc/helpers/locale');
var Base = require('syracuse-tablet/html/js/wc/controls/array/builderChartBase').BuilderChartBase;
var notifications = require('syracuse-tablet/html/js/wc/helpers/notifications');


var _getHeight = function(h) {
	if (h === "xsmall") return 50;
	if (h === "small") return 100;
	if (h === "medium") return 200;
	if (h === "large") return 400;
	if (h === "xlarge") return 700;
	return 200;
};

/**
 * HIGHCHART BUILDER
 */
exports.BuilderChart = utils.defineClass(
	function builderChart(control) {
		Base.call(this, control);
	}, Base, {
		initReuseProperties: function() {
			Base.prototype.initReuseProperties.call(this);
			this.resetProps(false);
		},
		destroy: function() {
			Base.prototype.destroy.call(this);
			this._$$chartSlot = this._$$catchEvt = null;
		},
		buildHtml: function(arrayData, $$content, buildOpts) {
			var self = this
			Base.prototype.buildHtml.call(this, arrayData, buildOpts);
			var article = this.$articleArray;
			// sometimes we need the height - see chartNeedsHeight
			this._height = _getHeight(this.control.article.$chartHeight);
			this._$$chartSlot = $('<div class="s-m-slot"/>').appendTo(this.control.$$content);
			this._setChartDataSet(arrayData)
				//this._dataset=this.control.arrayDataset.getData();
			if (this._dataset.length > 0) {
				//isAuthoringActive to allow the selection of the chart
				if (globals.isAuthoringActive()) {
					// Catch chart events
					this._$$catchEvt = $('<div class="s-m-catchevt"/>').appendTo(this.control.$$value);
				}
			} else {
				var title = this.control.prototype.getValueByPath("$cube.$title");
				if (title) {
					$('<div class="s-m-no-data-title"/>').appendTo(this._$$chartSlot).text(title);
				}
				$('<div class="s-m-no-data-label"/>').appendTo(this._$$chartSlot).text(locale.text("label.chart.nodata"));
				this._$$chartSlot.css("position", "relative");
			}
			// Create the header - Pagination if any
			this.refreshHeaderAndGesture();
			// in refesh context it's necessary to _displayChart because buildOnResize is not called
			//if (this.isRefreshContext(buildOpts)) {
			this._displayChart(false, true);
			//}
		},
		buildOnResize: function(arrayData, $$content, context) {
			//TODO:if (this.isDestroyed()) return;
			if (arrayData == null || arrayData.length === 0) {
				// It needs to count all charts even empty ones
				this._notifyChartCounter();
				return;
			}
			if (!$$content || !$$content.is(':visible')) {
				if (!this._subscribedExpand) {
					this._subscribedExpand = true;
					notifications.subscribe(this, ["sm.layout.expanded"]);
				}
			}
			if (this._layouytType == null) {
				// Needed here to access to all elements (getParentVignette)
				// We do that just once
				/* TODO if (this.control.isDisplayedInDetailPage()) {
					// in that case the chart is displayed in a 'detail's page
					this._layouytType = "stack";
				} else {
					// Chart in vignettes
					this._layouytType = this.control.controller.getParentDashboardType();
				}*/
			}
			this._displayChart(true, false);
		},
		_checkHeaderHeight: function(height) {
			if (this.control.builderHeader == null) {
				return height;
			}
			var $$header = this.control.builderHeader.$$elmt;
			if (!$$header.is(":visible")) {
				return height;
			}
			var hh = $$header.height();
			var phh = $$header ? 2 : 0; //+2 to display bottom border (???)
			if (height == null) {
				var css = {
					top: hh + "px",
					height: (this.control.$$content.height() - hh - phh) + "px"
				};
				this._$$chartSlot.css(css);
				if (this._$$catchEvt) {
					this._$$catchEvt.css(css);
				}
			} else {
				height = height - hh - phh;
			}
			return height;
		},
		_displayChart: function(updtScroller, isRefresh) {
			var self = this;
			if ( /*TODO : self.isDestroyed() || */ !self._$$chartSlot) return;
			// SetTimeout to improve UX when there are multiple charts to display (dashboard)
			setTimeout(function() {
				//TODO :if (self.isDestroyed()) return;
				var stackController, height;
				if (self._layouytType === "stack") {
					// relative is needed - set each time because aof refresh
					self._$$chartSlot.css("position", "relative");
					// in stack mode we need a height
					if (self.control.isDisplayedInDetailPage()) {
						stackController = self.control.controller;
						// Height set to device height
						height = stackController.getMainContentRect().height - 10;
					} else if (self.control.controller.isVignette) {
						stackController = self.control.controller.getParentVignette().controller;
						// Height always less or equal than the device height
						height = Math.min(self._height, stackController.getMainContentRect().height - 10);
					} else {
						throw new Error("Unexpected chart page context");
					}
				} else {
					// Parent's height
					height = null;
				}
				height = self._checkHeaderHeight(height);
				self.createChart(self._settingsFromValue(), self._$$chartSlot, {
					height: height,
					// Parent's width
					width: null,
					onComplete: function() {
						if (!isRefresh) {
							//#6974 Fix svg issue in chrome
							self._notifyChartCounter();
						}
					}
				});
				if (updtScroller && stackController) {
					// Needed to update the scroller
					notifications.notifyController("sm.scroller.update", stackController);
				}
			}, 10);
		},
		/**
		 * Parent control notifies the builder that it's going to clear the content
		 */
		notifClearContent: function(refresh) {
			this.clearChart();
		},
		notifLayoutExpanded: function() {
			this._displayChart(false, false);
		},
		/**
		 * #6974 When all the charts are loaded we force the browser to redraw (svg bug in chrome)
		 * We need to count the charts
		 */
		_notifyChartCounter: function() {
			return;
			//TODO 
			if (globals.getCurrentPage().countChartLoaded) {
				globals.getCurrentPage().countChartLoaded(this);
			}
		},
		_getChartDataSet: function() {
			return this._dataset || [];
		},
		_setChartDataSet: function(arrayData) {
			var chartDataset = [];
			if (arrayData && arrayData.length > 0) {
				arrayData.forEach(function(rowData) {
					chartDataset.push(rowData.getData());
				});
			} //TODO Manage link (drilldown ...)
			this._dataset = chartDataset;
		}
	});
