// syracuse-tablet/html/js/wc/controls/array/builderHeader.js
"use strict";

var utils = require('syracuse-tablet/html/js/wc/helpers/utils');
var FiltersSort = require('syracuse-tablet/html/js/wc/controls/array/filtersSort');

var _html = {
	header: '<header> \
			<nav class="s-m-stdheader"> \
				{{#if showPagination}} \
				<nav class="s-m-pagin"> \
					<ul class="pagination"> \
						<li> \
							<a class="s-m-link glyphicon glyphicon-step-backward disabled" href="#" \
								data-nevent \
								data-naction="pagination" \
								data-params="$first" \
								draggable="false"> \
							</a> \
						</li> \
						<li> \
							<a class="s-m-link glyphicon glyphicon-chevron-left disabled" href="#" \
								data-nevent \
								data-naction="pagination" \
								data-params="$previous" \
								draggable="false"> \
							</a> \
						</li> \
						<li> \
							<a class="s-m-link glyphicon glyphicon-chevron-right disabled" href="#" \
								data-nevent \
								data-naction="pagination" \
								data-params="$next" \
								draggable="false "> \
							</a> \
						</li> \
						<li> \
							<a class="s-m-link glyphicon glyphicon-step-forward disabled" href="#" \
								data-nevent \
								data-naction="pagination" \
								data-params="$last" \
								draggable="false"> \
							</a> \
						</li> \
					</ul> \
				</nav> \
				{{/if}} \
				{{#if showSearch}} \
				<nav class="s-m-search">\
				</nav>\
				{{/if}} \
				{{#if showSortFilter}} \
				<div class="s-m-filterSort" data-nevent data-naction="actionArray" data-params="panelfiltersort">\
					<i class="s-m-sorted" style="display:none"></i>\
					<i class="s-m-filtered fa fa-filter" style="display:none"></i>\
					<i class="fa fa-bars"></i>\
				</div>\
				{{/if}}\
				{{#if showFilters}}\
				<nav class="s-m-filterslist">\
				</nav>\
				{{/if}}\
			</nav> \
		</header>'
};

var _templates = {};
var _getHtml = function(name, ctx) {
	var tmpl = _templates[name];
	if (!tmpl) {
		tmpl = _templates[name] = Handlebars.compile(_html[name]);
	}
	return tmpl(ctx || {});
};

/**
 * Render array header
 */
exports.BuilderHeader = utils.defineClass(
	function builderHeader(control) {
		this.control = control;
	}, null, {

		destroy: function() {
			this.control = null;
			if (this.$$elmt) {
				this.$$elmt.remove()
			}
			this.$$elmt = null;
		},

		buildHtml: function() {
			if (this.$$elmt) {
				this.$$elmt.remove();
			}
			var options = {
				showPagination: true,
				showFilters: !this.control.isArrayChart() && this.control._filters.$filtersAllowed(),
				showSortFilter: !this.control.isArrayChart() && FiltersSort.filterSortAllowed(this.control),
				showSearch: !this.control.isArrayChart() // TODO : compute showsearch (authoring etc ...)
			}
			var html = _getHtml("header", options);
			this.$$elmt = $(html);
			if (this.control._search) {
				$("nav.s-m-search", this.$$elmt).append(this.control._search.getHtml())
			}
			if (this.control._filters) {
				$("nav.s-m-filterslist", this.$$elmt).append(this.control._filters.getHtml("filters"));
			}
			this.control.$$content.before(this.$$elmt);
			this.update();
			return this.$$elmt;
		},

		update: function() {
			if (this.control.paginator) {
				var links = this.control.paginator.pagingLinks;
				var self = this;
				links.forEach(function(link) {
					var $$link = $('[data-naction="pagination"][data-params="' + link + '"]', self.$$elmt);
					var ena = self.control.paginator.isEnabled(link);
					$$link.toggleClass("hidden", !ena);
					$$link.toggleClass("disabled", !ena);
				});
			}
		}

	});
