// syracuse-tablet/html/js/wc/controls/array/ctrlArrayTableRow.js
"use strict";

var Base = require('syracuse-tablet/html/js/wc/controls/ctrlSdataBase').CtrlSdataBase;
var utils = require('syracuse-tablet/html/js/wc/helpers/utils');
var uiUtils = require('syracuse-tablet/html/js/wc/ui/uiUtils');
var modules = require('syracuse-tablet/html/js/wc/common/modules');
var fontUtils = require('syracuse-tablet/html/js/wc/ui/fontUtils');
var builderBase = require('syracuse-tablet/html/js/wc/controls/array/builderBase');


/**
 * Control representing a single row of an array
 */
exports.CtrlArrayTableRow = utils.defineClass(
	function CtrlArrayTableRow(controller, article, prototype, options) {
		Base.call(this, controller, article, prototype, options);
		this.$properties = this.prototype.getValueByPath("$properties");
	}, Base, {
		destroy: function() {
			Base.prototype.destroy.call(this);
			this.$properties = null;
		},

		buildHtml: function(columnsInfo) {
			var classes = ["s-m-record"];
			if (this.options.rowIndex != null && this.options.rowIndex % 2 === 0) {
				classes.push("s-m-even");
			}
			var ctrlFactory = modules.get("ctrlFactory");

			Base.prototype.buildHtml.call(this, classes);
			if (this.options.rowDetailLinkDomAttrs) {
				this.$$elmt.attr(this.options.rowDetailLinkDomAttrs);
			}
			this.$$elmt.attr({
				"data-controller-id": this.controller.id
			});
			var self = this;
			columnsInfo.forEach(function(col, idx) {
				if (col.isProtoItem) {
					var ctrlCell = ctrlFactory.createControl(self.controller, col,
						self,
						self.page, {
							displayCtx: "table",
							noEdit: self.options.noEdit
						});
					if (ctrlCell) {
						var $$td = $(uiUtils.createDomElement("td", null, null, null, self.$$elmt));
						ctrlCell.set$$container($$td);
						ctrlCell.buildHtml();
					}
				} else {
					var text;
					switch (col.$bind) {
						case "rowStatus":
							text = builderBase.getHtmlStatus(col.statusInfo, self.controller.dataset)
							break;
						case "rowIndex":
							text = '<div class="s-m-control s-m-row-index"><span class="badge">' + (self.options.rowIndex + 1) + '</span></div>';
							break;
						case "rowDetail":
						case "rowDelete":
							text = '<div class="s-m-control s-m-icon"><a href="#" data-nevent data-naction="' + col.$bind + '" class="a"><span class="' + fontUtils.tableActionIcon(col.$bind) + '"></span></a></div>';
							break;
					}
					$("<td>" + (text || col.$bind) + "</td>").appendTo(self.$$elmt);
				}
			});
		},
		getRootElementTag: function() {
			return "tr";
		},

		_actLookupAccept: function() {
			var lookup = this.controller.root.lookupData; // Injected by lookup page so we can use it here
			var parentController = this.controller.findController(lookup.controllerId);
			var value = this.controller.getValue(lookup.$lookupKeyName);
			parentController.setValue(lookup.$bind, value, true);

			return modules.get("appController").App.goBack();
		},

		_actRowDetail: function() {
			this.controller.gotoUrl("child://SdataRowDetail", null, {
				isChild: true,
				childContext: {
					controllerId: this.controller.id
				}
			});
		},
		_actDeleteRow: function() {
			// Ask the parent (=> array to remove the row)
			this.parent.removeRow(this);
		}
	});
