// syracuse-tablet/html/js/wc/controls/array/ctrlArrayCardCell.js
"use strict";

var Base = require('syracuse-tablet/html/js/wc/controls/ctrlSdataBase').CtrlSdataBase;
var utils = require('syracuse-tablet/html/js/wc/helpers/utils');
var uiUtils = require('syracuse-tablet/html/js/wc/ui/uiUtils');
var modules = require('syracuse-tablet/html/js/wc/common/modules');
var fontUtils = require('syracuse-tablet/html/js/wc/ui/fontUtils');
var builderBase = require('syracuse-tablet/html/js/wc/controls/array/builderBase');
var SdataController = require("syracuse-tablet/html/js/wc/controllers/sdataController").SdataController;


/**
 * Control representing a single cell of a card array
 */
exports.CtrlArrayCardCell = utils.defineClass(
	function CtrlArrayCardCell(rowDataset, arrayController, article, prototype, options) {
		// first
		this.cellController = new SdataController(rowDataset, arrayController);
		// Second
		Base.call(this, arrayController, article, prototype, options);
	}, Base, {
		destroy: function() {
			Base.prototype.destroy.call(this);
			if (this.cellController) {
				this.cellController.destroy();
			}
			this.cellController = null;
		},

		createId: function(typeName) {
			return "rowid-" + this.getUuid();
		},

		buildHtml: function(columnsInfo) {
			var classes = ["s-m-record"];
			if (this.options.rowIndex != null && this.options.rowIndex % 2 === 0) {
				classes.push("s-m-even");
			}
			var ctrlFactory = modules.get("ctrlFactory");

			Base.prototype.buildHtml.call(this, classes);
			if (this.options.rowLinkDomAttrs) {
				this.$$elmt.attr(this.options.rowLinkDomAttrs);
			}
			this.$$elmt.attr({
				"data-controller-id": this.cellController.id
			});
		},

		getRootElementTag: function() {
			return "div";
		},

		onSdataLinkClicked: function($linkName, $linkBind) {
			this.setCellSelected();
			this.cellController.onSdataLinkClicked($linkName, $linkBind);
		},

		getUuid: function() {
			return this.cellController.dataset.json.$uuid;
		},

		setCellSelected: function(select) {
			select = select == null ? true : false;
			this.parent.$$content.find('.s-m-last-selected').removeClass("s-m-last-selected");
			if (select) {
				return this.$$elmt.addClass("s-m-last-selected");
			}
		}
	});
