// syracuse-ui/lib/fusion/article/fusionBarOld.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Box = require('syracuse-ui/lib/common/article/box').Box;

function FusionBarOld(){
}

exports.FusionBarOld = helpers.defineClass(FusionBarOld, Box, {
    drawBox: function(){
        this.$$item = $("<div class='s-fusion-bar'/>").appendTo(this.$$container);
        var $prototype = this.getArticle().$prototype;
        var $accordionId = document.controller.generateUUID();
        var $blocks = this.$item.$items.map(function($item){
            $item.$isTitleHidden = true;
            return {
                $category: "block",
                $XID: $item.$bind, // Necessary for Fusion
                $accordionId: $accordionId,
                $title: $item.$title || $prototype.$properties[$item.$bind].$title,
                $layout: {
                    $items: [$item]
                }
            }
        });
        $blocks[0].$opened = true;
        document.itemFactory.load(this.$$item, {
            $category: "section",
            $skin: "",
            $layout: {
                $items: $blocks
            }
        }, this);
    },
    dispose: function(){
        Box.prototype.dispose.call(this);
    }
});
