// syracuse-ui/lib/fusion/article/page.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var DesktopPage = require('syracuse-ui/lib/desktop/article/page').DesktopPage;
var syraUtil = require('syracuse-ui/lib/fusion/core/client/sap/util').Syra;
var FusionBar = require('./fusionBar').FusionBar;


function FusionPage(){
}

exports.FusionPage = helpers.defineClass(FusionPage, DesktopPage, {
    bind: function(item, $bind){
        var data, $fusionPageMeta = syraUtil.getFusionPageMeta(this);
        (this.boundFields[$bind] = this.boundFields[$bind] || []).push(item);
        if ($fusionPageMeta && $fusionPageMeta.winModel && (data = $fusionPageMeta.winModel.getDatasetValue($bind))) {
            item.setDataBind(data.val, data.rcd, data.meta);
            //item.setDataBind(store[$bind], store); // TODO : remove : syracuse way!
        }
        // TODO : idem for meta data!!!
    },
    onWindowResize: function(){
        if (this.$$sheetOverlay) {
            this.$$sheetOverlay.css({
                width: this.$$item.outerWidth(),
                height: this.$$item.outerHeight(true)
            });
        }
        if (this.fusionBar) {
            this.fusionBar.onWindowResize();
        }
    },
    selectSheet: function(selected){
        if (selected) {
            document.site.addEventListener("resize", this);
        }
        else {
            this.toggleSheetOverlay(false);
            if (document.site.removeEventListener) {
                document.site.removeEventListener("resize", this);
            }
        }
    },
    toggleSheetOverlay: function(show){
        var self = this;
        if (show) {
            if (!self.$$sheetOverlay) {
                document.site.$$body.append(self.$$sheetOverlay = $("<div class='s-fusion-sheet-overlay'/>"));
            }
            self.onWindowResize();
        }
        else {
            if (self.$$sheetOverlay) {
                self.$$sheetOverlay.remove();
                delete self.$$sheetOverlay;
            }
        }
    },
    drawBox: function(){
        DesktopPage.prototype.drawBox.call(this);
        if (this.fusionBar) {
            if (this.layoutContent.$layout.$layoutSubType == "25,75") {
                var cell = this.layoutContent.items[0].$$container[0];
                cell.style.display = "none";
                cell.style.width = "";
                cell = this.layoutContent.items[1].$$container[0];
                cell.style.width = "100%";
            }
        }
    },
    appendFusionBar: function($fusionBar){
        this.fusionBar = new FusionBar();
        this.fusionBar.load(this, $fusionBar);
    },
    dispose: function(){
        this.selectSheet(false);
        DesktopPage.prototype.dispose.call(this);
        syraUtil.deleteFusionPageMeta(this);
    },
    applyChange: function(newData){
        // TODO : surcharger a vide cette methode pour desactiver cette appel inutile en fusion????
        DesktopPage.prototype.applyChange.call(this);
    }
});
