// syracuse-ui/lib/fusion/article/sheetPage.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var DesktopPage = require('syracuse-ui/lib/article/page').DesktopPage;
var syraUtil = require('syracuse-ui/lib/fusion/core/client/sap/util').Syra;
var FusionBar = require('./fusionBar').FusionBar;

function SheetPage(){
}

exports.SheetPage = helpers.defineClass(SheetPage, DesktopPage, {
    loadBox: function(initData){
        if (!this.$item.$fusionBar) {
            var $rootItems = this.$item.$layout.$items;
            if ($rootItems && $rootItems.length == 2) {
                if ($rootItems[0].$items && $rootItems[0].$items[0].$category == "fusionBar") {
                    this.$item.$fusionBar = $rootItems[0].$items[0];
                    this.$item.$layout = $rootItems[1];
                }
            }
        }
        DesktopPage.prototype.loadBox.call(this, initData);
        if (this.$item.$fusionBar) {
            this.appendFusionBar(this.$item.$fusionBar);
        }
    },
    createFusionBar: function(){
        (this.fusionBar = new FusionBar()).load(this, this.$item.$fusionBar);
    },
    drawBox: function(){
        DesktopPage.prototype.drawBox.call(this);
        this.isFusionPage = true;
        this.$$item.attr("data-s-fusion-page", "1");
    },
    appendFusionBar: function($fusionBar){
        (this.fusionBar = new FusionBar()).load(this, $fusionBar);
    },
    bind: function(item, $bind){
        var data, $fusionPageMeta = syraUtil.getFusionPageMeta(this);
        (this.boundFields[$bind] = this.boundFields[$bind] || []).push(item);
        if ($fusionPageMeta && $fusionPageMeta.winModel && (data = $fusionPageMeta.winModel.getDatasetValue($bind, null, true))) {
            item.setDataBind(data.val, data.rcd, data.meta);
        }
        // TODO : idem for meta data!!!
    },
    onWindowResize: function(){
        if (this.isSelected && this.fusionBar) {
            this.fusionBar.onWindowResize();
        }
        DesktopPage.prototype.onWindowResize.call(this);
    },
    selectSheet: function(isSelected){
        if (this.isSelected = isSelected) {
            this.onWindowResize();
        }
    },
    dispose: function(){
        delete this.externalAdapter;
        syraUtil.deleteFusionPageMeta(this);
        DesktopPage.prototype.dispose.call(this);
    }
});
