// syracuse-ui/lib/tests/array/delta/deltaTests.js
"use strict";
exports.$menus = {
	$layout: {
		$items: [{
			$bind: "$isSelected"
		}, {
			$layoutType: "row",
			$items: [{
				$title: "update Title",
				$layout: {
					$items: [{
						$bind: "upadateTitleArray"
					}, {
						$bind: "upadateTitleColumns"
					}]
				}
			}, {
				$title: "show /hide",
				$layout: {
					$items: [{
						$bind: "hidexarray"
					}, {
						$bind: "showxarray"
					}, {
						$bind: "hidecolumn"
					}, {
						$bind: "showcolumn"
					}]
				}
			}]
		}, {
			$layoutType: "row",
			$items: [{
				$title: "$isRedOnly",
				$layout: {
					$items: [{
						$bind: "$isRedOnlyallarray"
					}, {
						$bind: "$isRedOnlyallarrayNo"
					}, {
						$bind: "$isRedOnlycolumn"
					}, {
						$bind: "$isRedOnlycolumnNo"
					}, {
						$bind: "$isRedOnlyrecord"
					}, {
						$bind: "$isRedOnlyrecordNo"
					}, {
						$bind: "$isRedOnlyrecords"
					}, {
						$bind: "$isRedOnlyrecordsNo"
					}]
				}
			}, {
				$title: "$isDisabled",
				$layout: {
					$items: [{
						$bind: "$isDisabledarray"
					}, {
						$bind: "$isDisabledcolumn"
					}, {
						$bind: "$isDisabledrecord"
					}, {
						$bind: "$isDisabledrecords"
					}]
				}
			}]
		}, {
			$title: "styles",
			$layout: {
				$items: [{
					$bind: "$fieldStylecolumn"
				}, {
					$bind: "$fieldStylerecords"
				}, {
					$bind: "$valueStyle"
				}]
			}
		}, {
			$layoutType: "row",
			$items: [{
				$title: "CRUD",
				$layout: {
					$items: [{
						$bind: "insert"
					}, {
						$bind: "updaterecord"
					}, {
						$bind: "updaterecords"
					}, {
						$bind: "reorder"
					}, {
						$bind: "deleterecord"
					}, {
						$bind: "deleterecords"
					}]
				}
			}, {
				$title: "diagnoses",
				$layout: {
					$items: [{
						$bind: "singlediagnose"
					}, {
						$bind: "multidiagnoses"
					}, {
						$bind: "cleandiagnoses"
					}]
				}
			}]
		}, {
			$title: "actionlinks",
			$layout: {
				$layoutType: "row",
				$items: [{
					$bind: "$linksnull"
				}, {
					$bind: "$linkslazyload"
				}, {
					$bind: "$links$actions"
				}, {
					$bind: "$linkParameterNull"
				}]
			}
		}, {
			$title: "capabilities",
			$layout: {
				$items: [{
					$bind: "capabilities"
				}, {
					$bind: "capabilitiesinsertdel"
				}, {
					$bind: "capabilitiesnone"
				}]
			}
		}]
	}
};

exports.$actions = {
	$isSelected: {
		$title: "$isSelected record 2",
		delta: {
			"xarray": [{
				"$isSelected": true,
				"$uuid": "record2",
				"$index": 2
			}]
		}
	},
	upadateTitleArray: {
		$title: "upadate title array",
		delta: {
			"$properties": {
				xarray: {
					$title: "new Title Default"
				}
			}
		}
	},
	upadateTitleColumns: {
		$title: "upadate xstring column Title",
		delta: {
			"$properties": {
				"xarray": {
					$item: {
						"$properties": {
							"xstring": {
								$title: "new xstring Default"
							}
						}
					}
				}
			}
		}
	},
	hidexarray: {
		$title: "hide xarray",
		delta: {
			"$properties": {
				"xarray": {
					"$isHidden": true
				}
			}
		}
	},
	showxarray: {
		$title: "showxarray",
		delta: {
			"$properties": {
				"xarray": {
					"$isHidden": false
				}
			}
		}
	},
	hidecolumn: {
		$title: "hide xchoice column",
		delta: {
			"$properties": {
				"xarray": {
					$item: {
						"$properties": {
							"xchoice": {
								"$isHidden": true
							}
						}
					}
				}
			}
		}
	},
	showcolumn: {
		$title: "show hidden xchoice column",
		delta: {
			"$properties": {
				"xarray": {
					$item: {
						"$properties": {
							"xchoice": {
								"$isHidden": false
							}
						}
					}
				}
			}
		}
	},
	$isRedOnlyallarray: {
		$title: "$isRedOnly all array",
		delta: {
			"$properties": {
				"xarray": {
					"$isReadOnly": true
				}
			}
		}
	},
	$isRedOnlyallarrayNo: {
		$title: "$isRedOnly all array no",
		delta: {
			"$properties": {
				"xarray": {
					"$isReadOnly": false
				}
			}
		}
	},
	$isRedOnlycolumn: {
		$title: "$isRedOnly xchoice xstring column",
		delta: {
			"$properties": {
				"xarray": {
					$item: {
						"$properties": {
							"xstring": {
								"$isReadOnly": true
							},
							"xchoice": {
								"$isReadOnly": true
							}
						}
					}
				}
			}
		}
	},
	$isRedOnlycolumnNo: {
		$title: "$isRedOnly xchoice xstring column no",
		delta: {
			"$properties": {
				"xarray": {
					$item: {
						"$properties": {
							"xstring": {
								"$isReadOnly": false
							},
							"xchoice": {
								"$isReadOnly": false
							}
						}
					}
				}
			}
		}
	},
	$isRedOnlyrecord: {
		$title: "$isRedOnly record 2",
		delta: {
			"xarray": [{
				"$isReadOnly": true,
				"$uuid": "record2",
				"$index": 2,
				"xstring": "record $isReadOnly"
			}]
		}
	},
	$isRedOnlyrecordNo: {
		$title: "$isRedOnly record 2 NO",
		delta: {
			"xarray": [{
				"$isReadOnly": false,
				"$uuid": "record2",
				"$index": 2,
				"xstring": "record $isReadOnly"
			}]
		}
	},
	$isRedOnlyrecordsNo: {
		$title: "$isRedOnly record3 record4 xstring xchoice NO",
		delta: {
			"xarray": [{
				"$uuid": "record0"
			}, {
				"$uuid": "record1"
			}, {
				"$uuid": "record2"
			}, {
				"$isReadOnly": false,
				"$uuid": "record3",
				"xstring": "record $isReadOnly"
			}, {
				"$properties": {
					"xstring": {
						"$isReadOnly": false
					},
					"xchoice": {
						"$isReadOnly": false
					}
				},
				"$uuid": "record4",
				"xstring": "xstring xchoice $isReadOnly"
			}, {
				"$uuid": "record5"
			}, {
				"$uuid": "record6",
				"xstring": "updated 6"
			}, {
				"$uuid": "record7",
				"xstring": "updated 7"
			}]
		}
	},
	$isRedOnlyrecords: {
		$title: "$isRedOnly record3 record4 xstring xchoice",
		delta: {
			"xarray": [{
				"$uuid": "record0"
			}, {
				"$uuid": "record1"
			}, {
				"$uuid": "record2"
			}, {
				"$isReadOnly": true,
				"$uuid": "record3",
				"xstring": "record $isReadOnly"
			}, {
				"$properties": {
					"xstring": {
						"$isReadOnly": true
					},
					"xchoice": {
						"$isReadOnly": true
					}
				},
				"$uuid": "record4",
				"xstring": "xstring xchoice $isReadOnly"
			}, {
				"$uuid": "record5"
			}, {
				"$uuid": "record6",
				"xstring": "updated 6"
			}, {
				"$uuid": "record7",
				"xstring": "updated 7"
			}]
		}
	},
	$isDisabledarray: {
		$title: "$isDisabled all array",
		delta: {
			"$properties": {
				"xarray": {
					"$isDisabled": true
				}
			}
		}
	},
	$isDisabledcolumn: {
		$title: "$isDisabled xchoice xstring column",
		delta: {
			"$properties": {
				"xarray": {
					$item: {
						"$properties": {
							"xstring": {
								"$isDisabled": true
							},
							"xchoice": {
								"$isDisabled": true
							}
						}
					}
				}
			}
		}
	},
	$isDisabledrecord: {
		$title: "$isDisabled record 2",
		delta: {
			"xarray": [{
				"$isDisabled": true,
				"$uuid": "record2",
				"$index": 2,
				"xstring": "record $isDisabled"
			}]
		}
	},
	$isDisabledrecords: {
		$title: "$isDisabled record3 record4 xstring xchoice",
		delta: {
			"xarray": [{
				"$uuid": "record0"
			}, {
				"$uuid": "record1"
			}, {
				"$uuid": "record2"
			}, {
				"$isDisabled": true,
				"$uuid": "record3",
				"xstring": "record $isDisabled"
			}, {
				"$properties": {
					"xstring": {
						"$isDisabled": true
					},
					"xchoice": {
						"$isDisabled": true
					}
				},
				"$uuid": "record4",
				"xstring": "xstring xchoice $isDisabled"
			}, {
				"$uuid": "record5"
			}, {
				"$uuid": "record6",
				"xstring": "updated 6"
			}, {
				"$uuid": "record7",
				"xstring": "updated 7"
			}]
		}
	},
	$fieldStylecolumn: {
		$title: "$fieldStyle xchoice xstring column",
		delta: {
			"$properties": {
				"xarray": {
					$item: {
						"$properties": {
							"xstring": {
								"$fieldStyle": "cst_sty_info"
							},
							"xchoice": {
								"$fieldStyle": "cst_sty_error"
							}
						}
					}
				}
			}
		}
	},
	$fieldStylerecords: {
		$title: "$fieldStyle record 2, $valueStyle record 3 record4 xstring xchoice",
		delta: {
			"xarray": [{
				"$uuid": "record0"
			}, {
				"$uuid": "record1"
			}, {
				"$fieldStyle": "cst_sty_info",
				"$uuid": "record2"
			}, {
				"$valueStyle": "cst_sty_error",
				"$uuid": "record3"
			}, {
				"$properties": {
					"xstring": {
						"$valueStyle": "cst_sty_warning"
					},
					"xchoice": {
						"$valueStyle": "cst_sty_veryhigh"
					}
				},
				"$uuid": "record4"
			}, {
				"$uuid": "record5"
			}, {
				"$uuid": "record6",
				"xstring": "updated 6"
			}, {
				"$uuid": "record7",
				"xstring": "updated 7"
			}]
		}
	},
	$valueStyle: {
		$title: "$fieldStyle $valueStyle $titleStyle default",
		delta: {
			"xarray": [{
				"$uuid": "record0",
				"$fieldStyle": "cst_sty_default",
				"$valueStyle": "cst_sty_default",
				"$titleStyle": "cst_sty_default"
			}, {
				"$uuid": "record1",
				"$fieldStyle": "cst_sty_default",
				"$valueStyle": "cst_sty_default",
				"$titleStyle": "cst_sty_default"
			}, {
				"$fieldStyle": "cst_sty_default",
				"$valueStyle": "cst_sty_default",
				"$titleStyle": "cst_sty_default",
				"$uuid": "record2"
			}, {
				"$fieldStyle": "",
				"$valueStyle": "",
				"$titleStyle": "",
				"$uuid": "record3"
			}, {
				"$fieldStyle": null,
				"$valueStyle": null,
				"$titleStyle": null,
				"$uuid": "record4"
			}, {
				"$fieldStyle": null,
				"$valueStyle": null,
				"$titleStyle": null,
				"$uuid": "record5"
			}, {
				"$uuid": "record6",
				"$fieldStyle": null,
				"$valueStyle": null,
				"$titleStyle": null
			}, {
				"$uuid": "record7",
				"$fieldStyle": null,
				"$valueStyle": null,
				"$titleStyle": null,
			}]
		}
	},
	insert: {
		$title: "insert 1",
		delta: {
			"xarray": [{
				"$uuid": "insert 1",
				"$index": 1,
				"xstring": "insert 1"
			}]
		}
	},
	updaterecord: {
		$title: "upadate record 2",
		delta: {
			"xarray": [{
				"$uuid": "record2",
				"$index": 2,
				"xstring": "updated 2"
			}]
		}
	},
	updaterecords: {
		$title: "upadate 2 3 6 7 ",
		delta: {
			"xarray": [{
				"$uuid": "record0"
			}, {
				"$uuid": "record1"
			}, {
				"$uuid": "record2",
				"xstring": "updated 2"
			}, {
				"$uuid": "record3",
				"xstring": "updated 3"
			}, {
				"$uuid": "record4"
			}, {
				"$uuid": "record5"
			}, {
				"$uuid": "record6",
				"xstring": "updated 6"
			}, {
				"$uuid": "record7",
				"xstring": "updated 7"
			}]
		}
	},
	reorder: {
		"$title": "reorder",
		"$id": "reorder",
		delta: {
			"xarray": []
		}
	},
	deleterecord: {
		"$title": "delete record2",
		delta: {
			"xarray": [{
				"$uuid": "record2",
				"$index": 2,
				"$isDeleted": true
			}]
		}
	},
	deleterecords: {
		"$title": "delete record3 record6 record7",
		"$id": "deleteDefault",
		delta: {
			"xarray": [{
				"$uuid": "record0"
			}, {
				"$uuid": "record1"
			}, {
				"$uuid": "record2"
			}, {
				"$uuid": "record4"
			}, {
				"$uuid": "record5"
			}]
		}
	},
	singlediagnose: {
		"$title": "single diagnose",
		delta: {
			"xarray": [{
				"$uuid": "record0",
				"$index": 0,
				"$properties": {
					"xstring": {
						"$diagnoses": [{
							"severity": "error",
							"message": "Error test  blalfoz"
						}]
					}
				}
			}]
		}
	},
	multidiagnoses: {
		"$title": "multi diagnoses",
		delta: {
			"$properties": {
				"xarray": {
					"$diagnoses": [{
						"severity": "error",
						"message": "Error test  blalfoz"
					}]
				}
			},
			"xarray": [{
				"$uuid": "record0",
				"$properties": {
					"xstring": {
						"$diagnoses": [{
							"severity": "error",
							"message": "Error test  blalfoz"
						}]
					},
					"xchoice": {
						"$diagnoses": [{
							"severity": "info",
							"message": "Info test"
						}]
					},
					"xdate": {
						"$diagnoses": [{
							"severity": "warning",
							"message": "Warning test"
						}]
					},
					"xboolean": {
						"$diagnoses": [{
							"severity": "error",
							"message": "Error test  blalfoz"
						}]
					},
					"xdecimal": {
						"$diagnoses": [{
							"severity": "error",
							"message": "Error test  blalfoz"
						}]
					}
				}
			}, {
				"$uuid": "record1"
			}, {
				"$uuid": "record2",
				"$properties": {
					"xstring": {
						"$diagnoses": [{
							"severity": "error",
							"message": "Error test  blalfoz"
						}]
					},
					"xchoice": {
						"$diagnoses": [{
							"severity": "warning",
							"message": "Warning test"
						}]
					}
				}
			}, {
				"$uuid": "record3"
			}, {
				"$uuid": "record4",
				"$properties": {
					"xstring": {
						"$diagnoses": [{
							"severity": "warning",
							"message": "Warning test"
						}]
					}
				}
			}, {
				"$uuid": "record5"
			}, {
				"$uuid": "record6",
				"$properties": {
					"xstring": {
						"$diagnoses": [{
							"severity": "error",
							"message": "Error test  blalfoz"
						}]
					},
					"xdecimal": {
						"$diagnoses": [{
							"severity": "error",
							"message": "Error test  blalfoz"
						}]
					}
				}
			}, {
				"$uuid": "record7"
			}]
		}
	},
	cleandiagnoses: {
		"$title": "clean diagnoses",
		delta: {
			"$properties": {
				"xarray": {
					"$diagnoses": []
				}
			},
			"xarray": [{
				"$uuid": "record0",
				"$properties": {
					"xstring": {
						"$diagnoses": []
					},
					"xchoice": {
						"$diagnoses": []
					},
					"xdate": {
						"$diagnoses": []
					},
					"xboolean": {
						"$diagnoses": []
					},
					"xdecimal": {
						"$diagnoses": []
					}
				}
			}, {
				"$uuid": "record1"
			}, {
				"$uuid": "record2",
				"$properties": {
					"xstring": {
						"$diagnoses": []
					},
					"xchoice": {
						"$diagnoses": []
					}
				}
			}, {
				"$uuid": "record3"
			}, {
				"$uuid": "record4",
				"$properties": {
					"xstring": {
						"$diagnoses": []
					}
				}
			}, {
				"$uuid": "record5"
			}, {
				"$uuid": "record6",
				"$properties": {
					"xstring": {
						"$diagnoses": []
					},
					"xdecimal": {
						"$diagnoses": []
					}
				}
			}, {
				"$uuid": "record7"
			}]
		}
	},
	$linksnull: {
		"$title": "$links = null",
		delta: {
			"$properties": {
				"xarray": {
					"$links": null
				}
			},
			"xarray": [{
				"$uuid": "record0"
			}, {
				"$uuid": "record1",
				"$properties": {
					"xstring": {
						"$links": null
					},
					"xchoice": {
						"$links": null
					},
					"xdate": {
						"$links": null
					},
					"xboolean": {
						"$links": null
					},
					"xdecimal": {
						"$links": null
					}
				},
				"$links": null
			}, {
				"$uuid": "record2"
			}, {
				"$uuid": "record3",
				"$properties": {
					"xstring": {
						"$links": null
					},
					"xchoice": {
						"$links": null
					},
					"xdate": {
						"$links": null
					},
					"xboolean": {
						"$links": null
					},
					"xdecimal": {
						"$links": null
					}
				},
				"$links": null
			}, {
				"$uuid": "record4"
			}, {
				"$uuid": "record5",
				"$properties": {
					"xstring": {
						"$links": null
					},
					"xchoice": {
						"$links": null
					},
					"xdate": {
						"$links": null
					},
					"xboolean": {
						"$links": null
					},
					"xdecimal": {
						"$links": null
					}
				},
				"$links": null
			}, {
				"$uuid": "record6"
			}, {
				"$uuid": "record7",
				"$properties": {
					"xstring": {
						"$links": null
					},
					"xchoice": {
						"$links": null
					},
					"xdate": {
						"$links": null
					},
					"xboolean": {
						"$links": null
					},
					"xdecimal": {
						"$links": null
					}
				},
				"xdecimal": 318178,
				"$links": null
			}]
		}
	},
	$linkslazyload: {
		"$title": "$lazyload",
		delta: {
			"$properties": {
				"xarray": {
					"$links": {
						"$lazyload": {}
					}
				}
			},
			"xarray": [{
				"$uuid": "record0"
			}, {
				"$uuid": "record1",
				"$properties": {
					"xstring": {
						"$links": {
							"$lazyload": {}
						}
					},
					"xchoice": {
						"$links": {
							"$lazyload": {}
						}
					},
					"xdate": {
						"$links": {
							"$lazyload": {}
						}
					},
					"xboolean": {
						"$links": {
							"$lazyload": {}
						}
					},
					"xdecimal": {
						"$links": {
							"$lazyload": {}
						}
					}
				},
				"$links": {
					"$lazyload": {}
				}
			}, {
				"$uuid": "record2"
			}, {
				"$uuid": "record3",
				"$properties": {
					"xstring": {
						"$links": {
							"$lazyload": {}
						}
					},
					"xchoice": {
						"$links": {
							"$lazyload": {}
						}
					},
					"xdate": {
						"$links": {
							"$lazyload": {}
						}
					},
					"xboolean": {
						"$links": {
							"$lazyload": {}
						}
					},
					"xdecimal": {
						"$links": {
							"$lazyload": {}
						}
					}
				},
				"$links": {
					"$lazyload": {}
				}
			}, {
				"$uuid": "record4"
			}, {
				"$uuid": "record5",
				"$properties": {
					"xstring": {
						"$links": {
							"$lazyload": {}
						}
					},
					"xchoice": {
						"$links": {
							"$lazyload": {}
						}
					},
					"xdate": {
						"$links": {
							"$lazyload": {}
						}
					},
					"xboolean": {
						"$links": {
							"$lazyload": {}
						}
					},
					"xdecimal": {
						"$links": {
							"$lazyload": {}
						}
					}
				},
				"$links": {
					"$lazyload": {}
				}
			}, {
				"$uuid": "record6"
			}, {
				"$uuid": "record7",
				"$properties": {
					"xstring": {
						"$links": {
							"$lazyload": {}
						}
					},
					"xchoice": {
						"$links": {
							"$lazyload": {}
						}
					},
					"xdate": {
						"$links": {
							"$lazyload": {}
						}
					},
					"xboolean": {
						"$links": {
							"$lazyload": {}
						}
					},
					"xdecimal": {
						"$links": {
							"$lazyload": {}
						}
					}
				},
				"$links": {
					"$lazyload": {}
				}
			}]
		}
	},
	$linkParameterNull: {
		"$title": "Parameter null",
		delta: {
			"$properties": {
				"xarray": {
					"$links": {
						"$linksFormParameters": null
					}
				}
			},
			"xarray": [{
				"$uuid": "record0"
			}, {
				"$uuid": "record1",
				"$properties": {
					"xstring": {
						"$links": {
							"$linksFormParameters": null
						}
					},
					"xchoice": {
						"$links": {
							"$linksFormParameters": null
						}
					},
					"xdate": {
						"$links": {
							"$linksFormParameters": null
						}
					},
					"xboolean": {
						"$links": {
							"$linksFormParameters": null
						}
					},
					"xdecimal": {
						"$links": {
							"$linksFormParameters": null
						}
					}
				},
				"$links": {
					"$linksFormParameters": null
				}
			}, {
				"$uuid": "record2"
			}, {
				"$uuid": "record3",
				"$properties": {
					"xstring": {
						"$links": {
							"$linksFormParameters": null
						}
					},
					"xchoice": {
						"$links": {
							"$linksFormParameters": null
						}
					},
					"xdate": {
						"$links": {
							"$linksFormParameters": null
						}
					},
					"xboolean": {
						"$links": {
							"$linksFormParameters": null
						}
					},
					"xdecimal": {
						"$links": {
							"$linksFormParameters": null
						}
					}
				},
				"$links": {
					"$linksFormParameters": null
				}
			}, {
				"$uuid": "record4"
			}, {
				"$uuid": "record5",
				"$properties": {
					"xstring": {
						"$links": {
							"$linksFormParameters": null
						}
					},
					"xchoice": {
						"$links": {
							"$linksFormParameters": null
						}
					},
					"xdate": {
						"$links": {
							"$linksFormParameters": null
						}
					},
					"xboolean": {
						"$links": {
							"$linksFormParameters": null
						}
					},
					"xdecimal": {
						"$links": {
							"$linksFormParameters": null
						}
					}
				},
				"$links": {
					"$linksFormParameters": null
				}
			}, {
				"$uuid": "record6"
			}, {
				"$uuid": "record7",
				"$properties": {
					"xstring": {
						"$links": {
							"$linksFormParameters": null
						}
					},
					"xchoice": {
						"$links": {
							"$linksFormParameters": null
						}
					},
					"xdate": {
						"$links": {
							"$linksFormParameters": null
						}
					},
					"xboolean": {
						"$links": {
							"$linksFormParameters": null
						}
					},
					"xdecimal": {
						"$links": {
							"$linksFormParameters": null
						}
					}
				},
				"$links": {
					"$linksFormParameters": null
				}
			}]
		}
	},
	$links$actions: {
		"$title": "$links ans $actions",
		delta: {
			"$properties": {
				"xarray": {
					"$actions": {
						"$save": {
							"$title": "Save"
						}
					},
					"$links": {
						"google": {
							"$title": "Google",
							"$url": "http://www.google.fr/"
						},
						"uioverview": {
							"$title": "UI Overview",
							"$url": "?representation=s-uitest-fields.$test"
						},
						"$details": {
							"$type": "application/json;vnd.sage=syracuse",
							"$title": "Detail",
							"$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
						},
						"$linksFormParameters": {
							"$title": "Parameters",
							"$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
							"$parameters": {
								"param1": "xstring",
								"param2": "value"
							}
						}
					}
				}
			},
			"xarray": [{
				"$uuid": "record0"
			}, {
				"$uuid": "record1",
				"$properties": {
					"xstring": {
						"$actions": {
							"$save": {
								"$title": "Save"
							}
						},
						"$links": {
							"google": {
								"$title": "Google",
								"$url": "http://www.google.fr/"
							},
							"uioverview": {
								"$title": "UI Overview",
								"$url": "?representation=s-uitest-fields.$test"
							},
							"$details": {
								"$type": "application/json;vnd.sage=syracuse",
								"$title": "Detail",
								"$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
							},
							"$linksFormParameters": {
								"$title": "Parameters",
								"$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
								"$parameters": {
									"param1": "xstring",
									"param2": "value"
								}
							}
						}
					},
					"xchoice": {
						"$actions": {
							"$save": {
								"$title": "Save"
							}
						},
						"$links": {
							"google": {
								"$title": "Google",
								"$url": "http://www.google.fr/"
							},
							"uioverview": {
								"$title": "UI Overview",
								"$url": "?representation=s-uitest-fields.$test"
							},
							"$details": {
								"$type": "application/json;vnd.sage=syracuse",
								"$title": "Detail",
								"$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
							},
							"$linksFormParameters": {
								"$title": "Parameters",
								"$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
								"$parameters": {
									"param1": "xstring",
									"param2": "value"
								}
							}
						}
					},
					"xdate": {
						"$actions": {
							"$save": {
								"$title": "Save"
							}
						},
						"$links": {
							"google": {
								"$title": "Google",
								"$url": "http://www.google.fr/"
							},
							"uioverview": {
								"$title": "UI Overview",
								"$url": "?representation=s-uitest-fields.$test"
							},
							"$details": {
								"$type": "application/json;vnd.sage=syracuse",
								"$title": "Detail",
								"$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
							},
							"$linksFormParameters": {
								"$title": "Parameters",
								"$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
								"$parameters": {
									"param1": "xstring",
									"param2": "value"
								}
							}
						}
					},
					"xboolean": {
						"$actions": {
							"$save": {
								"$title": "Save"
							}
						},
						"$links": {
							"google": {
								"$title": "Google",
								"$url": "http://www.google.fr/"
							},
							"uioverview": {
								"$title": "UI Overview",
								"$url": "?representation=s-uitest-fields.$test"
							},
							"$details": {
								"$type": "application/json;vnd.sage=syracuse",
								"$title": "Detail",
								"$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
							},
							"$linksFormParameters": {
								"$title": "Parameters",
								"$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
								"$parameters": {
									"param1": "xstring",
									"param2": "value"
								}
							}
						}
					},
					"xdecimal": {
						"$actions": {
							"$save": {
								"$title": "Save"
							}
						},
						"$links": {
							"google": {
								"$title": "Google",
								"$url": "http://www.google.fr/"
							},
							"uioverview": {
								"$title": "UI Overview",
								"$url": "?representation=s-uitest-fields.$test"
							},
							"$details": {
								"$type": "application/json;vnd.sage=syracuse",
								"$title": "Detail",
								"$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
							},
							"$linksFormParameters": {
								"$title": "Parameters",
								"$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
								"$parameters": {
									"param1": "xstring",
									"param2": "value"
								}
							}
						}
					}
				},
				"$actions": {
					"$save": {
						"$title": "Save"
					}
				},
				"$links": {
					"google": {
						"$title": "Google",
						"$url": "http://www.google.fr/"
					},
					"uioverview": {
						"$title": "UI Overview",
						"$url": "?representation=s-uitest-fields.$test"
					},
					"$details": {
						"$type": "application/json;vnd.sage=syracuse",
						"$title": "Detail",
						"$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
					},
					"$linksFormParameters": {
						"$title": "Parameters",
						"$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
						"$parameters": {
							"param1": "xstring",
							"param2": "value"
						}
					}
				}
			}, {
				"$uuid": "record2"
			}, {
				"$uuid": "record3",
				"$properties": {
					"xstring": {
						"$actions": {
							"$save": {
								"$title": "Save"
							}
						},
						"$links": {
							"google": {
								"$title": "Google",
								"$url": "http://www.google.fr/"
							},
							"uioverview": {
								"$title": "UI Overview",
								"$url": "?representation=s-uitest-fields.$test"
							},
							"$details": {
								"$type": "application/json;vnd.sage=syracuse",
								"$title": "Detail",
								"$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
							},
							"$linksFormParameters": {
								"$title": "Parameters",
								"$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
								"$parameters": {
									"param1": "xstring",
									"param2": "value"
								}
							}
						}
					},
					"xchoice": {
						"$actions": {
							"$save": {
								"$title": "Save"
							}
						},
						"$links": {
							"google": {
								"$title": "Google",
								"$url": "http://www.google.fr/"
							},
							"uioverview": {
								"$title": "UI Overview",
								"$url": "?representation=s-uitest-fields.$test"
							},
							"$details": {
								"$type": "application/json;vnd.sage=syracuse",
								"$title": "Detail",
								"$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
							},
							"$linksFormParameters": {
								"$title": "Parameters",
								"$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
								"$parameters": {
									"param1": "xstring",
									"param2": "value"
								}
							}
						}
					},
					"xdate": {
						"$actions": {
							"$save": {
								"$title": "Save"
							}
						},
						"$links": {
							"google": {
								"$title": "Google",
								"$url": "http://www.google.fr/"
							},
							"uioverview": {
								"$title": "UI Overview",
								"$url": "?representation=s-uitest-fields.$test"
							},
							"$details": {
								"$type": "application/json;vnd.sage=syracuse",
								"$title": "Detail",
								"$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
							},
							"$linksFormParameters": {
								"$title": "Parameters",
								"$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
								"$parameters": {
									"param1": "xstring",
									"param2": "value"
								}
							}
						}
					},
					"xboolean": {
						"$actions": {
							"$save": {
								"$title": "Save"
							}
						},
						"$links": {
							"google": {
								"$title": "Google",
								"$url": "http://www.google.fr/"
							},
							"uioverview": {
								"$title": "UI Overview",
								"$url": "?representation=s-uitest-fields.$test"
							},
							"$details": {
								"$type": "application/json;vnd.sage=syracuse",
								"$title": "Detail",
								"$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
							},
							"$linksFormParameters": {
								"$title": "Parameters",
								"$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
								"$parameters": {
									"param1": "xstring",
									"param2": "value"
								}
							}
						}
					},
					"xdecimal": {
						"$actions": {
							"$save": {
								"$title": "Save"
							}
						},
						"$links": {
							"google": {
								"$title": "Google",
								"$url": "http://www.google.fr/"
							},
							"uioverview": {
								"$title": "UI Overview",
								"$url": "?representation=s-uitest-fields.$test"
							},
							"$details": {
								"$type": "application/json;vnd.sage=syracuse",
								"$title": "Detail",
								"$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
							},
							"$linksFormParameters": {
								"$title": "Parameters",
								"$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
								"$parameters": {
									"param1": "xstring",
									"param2": "value"
								}
							}
						}
					}
				},
				"$actions": {
					"$save": {
						"$title": "Save"
					}
				},
				"$links": {
					"google": {
						"$title": "Google",
						"$url": "http://www.google.fr/"
					},
					"uioverview": {
						"$title": "UI Overview",
						"$url": "?representation=s-uitest-fields.$test"
					},
					"$details": {
						"$type": "application/json;vnd.sage=syracuse",
						"$title": "Detail",
						"$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
					},
					"$linksFormParameters": {
						"$title": "Parameters",
						"$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
						"$parameters": {
							"param1": "xstring",
							"param2": "value"
						}
					}
				}
			}, {
				"$uuid": "record4"
			}, {
				"$uuid": "record5",
				"$properties": {
					"xstring": {
						"$actions": {
							"$save": {
								"$title": "Save"
							}
						},
						"$links": {
							"google": {
								"$title": "Google",
								"$url": "http://www.google.fr/"
							},
							"uioverview": {
								"$title": "UI Overview",
								"$url": "?representation=s-uitest-fields.$test"
							},
							"$details": {
								"$type": "application/json;vnd.sage=syracuse",
								"$title": "Detail",
								"$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
							},
							"$linksFormParameters": {
								"$title": "Parameters",
								"$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
								"$parameters": {
									"param1": "xstring",
									"param2": "value"
								}
							}
						}
					},
					"xchoice": {
						"$actions": {
							"$save": {
								"$title": "Save"
							}
						},
						"$links": {
							"google": {
								"$title": "Google",
								"$url": "http://www.google.fr/"
							},
							"uioverview": {
								"$title": "UI Overview",
								"$url": "?representation=s-uitest-fields.$test"
							},
							"$details": {
								"$type": "application/json;vnd.sage=syracuse",
								"$title": "Detail",
								"$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
							},
							"$linksFormParameters": {
								"$title": "Parameters",
								"$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
								"$parameters": {
									"param1": "xstring",
									"param2": "value"
								}
							}
						}
					},
					"xdate": {
						"$actions": {
							"$save": {
								"$title": "Save"
							}
						},
						"$links": {
							"google": {
								"$title": "Google",
								"$url": "http://www.google.fr/"
							},
							"uioverview": {
								"$title": "UI Overview",
								"$url": "?representation=s-uitest-fields.$test"
							},
							"$details": {
								"$type": "application/json;vnd.sage=syracuse",
								"$title": "Detail",
								"$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
							},
							"$linksFormParameters": {
								"$title": "Parameters",
								"$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
								"$parameters": {
									"param1": "xstring",
									"param2": "value"
								}
							}
						}
					},
					"xboolean": {
						"$actions": {
							"$save": {
								"$title": "Save"
							}
						},
						"$links": {
							"google": {
								"$title": "Google",
								"$url": "http://www.google.fr/"
							},
							"uioverview": {
								"$title": "UI Overview",
								"$url": "?representation=s-uitest-fields.$test"
							},
							"$details": {
								"$type": "application/json;vnd.sage=syracuse",
								"$title": "Detail",
								"$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
							},
							"$linksFormParameters": {
								"$title": "Parameters",
								"$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
								"$parameters": {
									"param1": "xstring",
									"param2": "value"
								}
							}
						}
					},
					"xdecimal": {
						"$actions": {
							"$save": {
								"$title": "Save"
							}
						},
						"$links": {
							"google": {
								"$title": "Google",
								"$url": "http://www.google.fr/"
							},
							"uioverview": {
								"$title": "UI Overview",
								"$url": "?representation=s-uitest-fields.$test"
							},
							"$details": {
								"$type": "application/json;vnd.sage=syracuse",
								"$title": "Detail",
								"$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
							},
							"$linksFormParameters": {
								"$title": "Parameters",
								"$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
								"$parameters": {
									"param1": "xstring",
									"param2": "value"
								}
							}
						}
					}
				},
				"$actions": {
					"$save": {
						"$title": "Save"
					}
				},
				"$links": {
					"google": {
						"$title": "Google",
						"$url": "http://www.google.fr/"
					},
					"uioverview": {
						"$title": "UI Overview",
						"$url": "?representation=s-uitest-fields.$test"
					},
					"$details": {
						"$type": "application/json;vnd.sage=syracuse",
						"$title": "Detail",
						"$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
					},
					"$linksFormParameters": {
						"$title": "Parameters",
						"$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
						"$parameters": {
							"param1": "xstring",
							"param2": "value"
						}
					}
				}
			}, {
				"$uuid": "record6"
			}, {
				"$uuid": "record7",
				"$properties": {
					"xstring": {
						"$actions": {
							"$save": {
								"$title": "Save"
							}
						},
						"$links": {
							"google": {
								"$title": "Google",
								"$url": "http://www.google.fr/"
							},
							"uioverview": {
								"$title": "UI Overview",
								"$url": "?representation=s-uitest-fields.$test"
							},
							"$details": {
								"$type": "application/json;vnd.sage=syracuse",
								"$title": "Detail",
								"$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
							},
							"$linksFormParameters": {
								"$title": "Parameters",
								"$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
								"$parameters": {
									"param1": "xstring",
									"param2": "value"
								}
							}
						}
					},
					"xchoice": {
						"$actions": {
							"$save": {
								"$title": "Save"
							}
						},
						"$links": {
							"google": {
								"$title": "Google",
								"$url": "http://www.google.fr/"
							},
							"uioverview": {
								"$title": "UI Overview",
								"$url": "?representation=s-uitest-fields.$test"
							},
							"$details": {
								"$type": "application/json;vnd.sage=syracuse",
								"$title": "Detail",
								"$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
							},
							"$linksFormParameters": {
								"$title": "Parameters",
								"$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
								"$parameters": {
									"param1": "xstring",
									"param2": "value"
								}
							}
						}
					},
					"xdate": {
						"$actions": {
							"$save": {
								"$title": "Save"
							}
						},
						"$links": {
							"google": {
								"$title": "Google",
								"$url": "http://www.google.fr/"
							},
							"uioverview": {
								"$title": "UI Overview",
								"$url": "?representation=s-uitest-fields.$test"
							},
							"$details": {
								"$type": "application/json;vnd.sage=syracuse",
								"$title": "Detail",
								"$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
							},
							"$linksFormParameters": {
								"$title": "Parameters",
								"$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
								"$parameters": {
									"param1": "xstring",
									"param2": "value"
								}
							}
						}
					},
					"xboolean": {
						"$actions": {
							"$save": {
								"$title": "Save"
							}
						},
						"$links": {
							"google": {
								"$title": "Google",
								"$url": "http://www.google.fr/"
							},
							"uioverview": {
								"$title": "UI Overview",
								"$url": "?representation=s-uitest-fields.$test"
							},
							"$details": {
								"$type": "application/json;vnd.sage=syracuse",
								"$title": "Detail",
								"$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
							},
							"$linksFormParameters": {
								"$title": "Parameters",
								"$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
								"$parameters": {
									"param1": "xstring",
									"param2": "value"
								}
							}
						}
					},
					"xdecimal": {
						"$actions": {
							"$save": {
								"$title": "Save"
							}
						},
						"$links": {
							"google": {
								"$title": "Google",
								"$url": "http://www.google.fr/"
							},
							"uioverview": {
								"$title": "UI Overview",
								"$url": "?representation=s-uitest-fields.$test"
							},
							"$details": {
								"$type": "application/json;vnd.sage=syracuse",
								"$title": "Detail",
								"$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
							},
							"$linksFormParameters": {
								"$title": "Parameters",
								"$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
								"$parameters": {
									"param1": "xstring",
									"param2": "value"
								}
							}
						}
					}
				},
				"$links": {
					"google": {
						"$title": "Google",
						"$url": "http://www.google.fr/"
					},
					"uioverview": {
						"$title": "UI Overview",
						"$url": "?representation=s-uitest-fields.$test"
					},
					"$details": {
						"$type": "application/json;vnd.sage=syracuse",
						"$title": "Detail",
						"$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
					},
					"$linksFormParameters": {
						"$title": "Parameters",
						"$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
						"$parameters": {
							"param1": "xstring",
							"param2": "value"
						}
					}
				},
				"$actions": {
					"$save": {
						"$title": "Save"
					}
				}
			}]
		}
	},
	capabilities: {
		$title: "capabilities insert,append,delete,reorder",
		delta: {
			"$properties": {
				"xarray": {
					"$capabilities": "insert,append,delete,reorder"
				}
			}
		}
	},
	capabilitiesinsertdel: {
		$title: "capabilities insert,delete,deleteAll",
		delta: {
			"$properties": {
				"xarray": {
					"$capabilities": "insert,delete,deleteAll"
				}
			}
		}
	},
	capabilitiesnone: {
		$title: "capabilities none",
		delta: {
			"$properties": {
				"xarray": {
					"$capabilities": ""
				}
			}
		}
	}
};
