// syracuse-ui/lib/tests/array/delta/partialDeltaTests.js
exports.cases = [{
    $title: "upadate title array",
    delta: {
        $isPartialDelta: true,
        $properties: {
            xarray: {
                $title: "new Title Partial"
            }
        }
    }
}, {
    $title: "upadate xstring column Title",
    delta: {
        $isPartialDelta: true,
        $properties: {
            xarray: {
                $item: {
                    $properties: {
                        xstring: {
                            $title: "new xstring Partial"
                        }
                    }
                }
            }
        }
    }
}, {
    $title: "hide xarray",
    delta: {
        $isPartialDelta: true,
        $properties: {
            xarray: {
                $isHidden: true
            }
        }
    }
}, {
    $title: "show hidden xarray",
    delta: {
        $isPartialDelta: true,
        $properties: {
            xarray: {
                $isHidden: false
            }
        }
    }
}, {
    $title: "hide xchoice column",
    delta: {
        $isPartialDelta: true,
        $properties: {
            xarray: {
                $item: {
                    $properties: {
                        xchoice: {
                            $isHidden: true
                        }
                    }
                }
            }
        }
    }
}, {
    $title: "show hidden xchoice column",
    delta: {
        $isPartialDelta: true,
        $properties: {
            xarray: {
                $item: {
                    $properties: {
                        xchoice: {
                            $isHidden: false
                        }
                    }
                }
            }
        }
    }
}, {
    $title: "$isReadOnly all array",
    delta: {
        $isPartialDelta: true,
        "$properties": {
            "xarray": {
                "$isReadOnly": true
            }
        }
    }
}, {
    $title: "$isReadOnly xchoice xstring column",
    delta: {
        $isPartialDelta: true,
        "$properties": {
            "xarray": {
                $item: {
                    "$properties": {
                        "xstring": {
                            "$isReadOnly": true
                        },
                        "xchoice": {
                            "$isReadOnly": true
                        }
                    }
                }
            }
        }
    }
}, {
    $title: "$isReadOnly record 2",
    delta: {
        $isPartialDelta: true,
        "xarray": [{
            "$isReadOnly": true,
            "$uuid": "record2",
            "$index": 2,
            "xstring": "record $isReadOnly"
        }]
    }
}, {
    $title: "$isReadOnly record3 record4 xstring xchoice",
    delta: {
        $isPartialDelta: true,
        "xarray": [{
            "$isReadOnly": true,
            "$index": 3,
            "$uuid": "record3",
            "xstring": "record $isReadOnly"
        }, {
            "$properties": {
                "xstring": {
                    "$isReadOnly": true
                },
                "xchoice": {
                    "$isReadOnly": true
                }
            },
            "$index": 4,
            "$uuid": "record4",
            "xstring": "xstring xchoice $isReadOnly"
        }]
    }
}, {
    $title: "$isDisabled all array",
    delta: {
        $isPartialDelta: true,
        "$properties": {
            "xarray": {
                "$isDisabled": true
            }
        }
    }
}, {
    $title: "$isDisabled xchoice xstring column",
    delta: {
        $isPartialDelta: true,
        "$properties": {
            "xarray": {
                $item: {
                    "$properties": {
                        "xstring": {
                            "$isDisabled": true
                        },
                        "xchoice": {
                            "$isDisabled": true
                        }
                    }
                }
            }
        }
    }
}, {
    $title: "$isDisabled record 2",
    delta: {
        $isPartialDelta: true,
        "xarray": [{
            "$isDisabled": true,
            "$uuid": "record2",
            "$index": 2,
            "xstring": "record $isDisabled"
        }]
    }
}, {
    $title: "$isDisabled record3 record4 xstring xchoice",
    delta: {
        $isPartialDelta: true,
        "xarray": [{
            "$isDisabled": true,
            "$index": 3,
            "$uuid": "record3",
            "xstring": "record $isDisabled"
        }, {
            "$properties": {
                "xstring": {
                    "$isDisabled": true
                },
                "xchoice": {
                    "$isDisabled": true
                }
            },
            "$index": 4,
            "$uuid": "record4",
            "xstring": "xstring xchoice $isDisabled"
        }]
    }
}, {
    $title: "$fieldStyle xchoice xstring column",
    delta: {
        $isPartialDelta: true,
        "$properties": {
            "xarray": {
                $item: {
                    "$properties": {
                        "xstring": {
                            "$fieldStyle": "font-style:bold; color: #FFFFFF;background-color:#FF80C0; background-color: #2E6AC2;background-image: linear-gradient(rgba(255, 255, 255, 0.1), rgba(255, 255, 255, 0.05) 49%, rgba(0, 0, 0, 0.05) 51%, rgba(0, 0, 0, 0.1));"
                        },
                        "xchoice": {
                            "$fieldStyle": "font-style:italic;background-color:#EA4C89; color: #FFFFFF;"
                        }
                    }
                }
            }
        }
    }
}, {
    $title: "$fieldStyle record 2, $valueStyle record 3 record4 xstring xchoice",
    delta: {
        $isPartialDelta: true,
        "xarray": [{
            "$fieldStyle": "font-style:bold; color: #FFFFFF;background-color:#FF80C0; background-color: #2E6AC2;background-image: linear-gradient(rgba(255, 255, 255, 0.1), rgba(255, 255, 255, 0.05) 49%, rgba(0, 0, 0, 0.05) 51%, rgba(0, 0, 0, 0.1));",
            "$uuid": "record2",
            "$index": 2
        }, {
            "$valueStyle": "font-style:italic;background-color:#EA4C89; color: #FFFFFF;",
            "$uuid": "record3",
            "$index": 3
        }, {
            "$properties": {
                "xstring": {
                    "$valueStyle": "font-style:bold;background-color:#FFAD1D; color: #030037;"
                },
                "xchoice": {
                    "$valueStyle": "font-style:bold;background-color:#030037; color: #FFAD1D;"
                }
            },
            "$index": 4,
            "$uuid": "record4"
        }]
    }
}, {
    $title: "insert 1",
    delta: {
        $isPartialDelta: true,
        "xarray": [{
            "$uuid": "insert 1",
            "$index": 1,
            "xstring": "insert 1"
        }]
    }
}, {
    $title: "insert 2",
    delta: {
        $isPartialDelta: true,
        "xarray": [{
            "$uuid": "insert 1",
            "$index": 1,
            "xstring": "insert 1"
        }, {
            "$uuid": "insert 5",
            "$index": 5,
            "xstring": "insert 5"
        }]
    }
}, {
    $title: "upd 2 4",
    delta: {
        $isPartialDelta: true,
        "xarray": [{
            "$uuid": "record2",
            "$index": 1,
            "xstring": "updated 2",
            "$properties": {
                "xstring": {
                    "$fieldStyle": "color:blue;"
                }
            }
        }, {
            "$uuid": "record 4",
            "$index": 3,
            "xstring": "updated 4",
            "$properties": {
                "xstring": {
                    "$fieldStyle": "color:red;",
                    "$diagnoses": [{
                        message: "test erreur"
                    }]
                }
            }
        }, {
            "$uuid": "record7",
            "$index": 6,
            "xstring": "updated 7",
            "$properties": {
                "xstring": {
                    "$fieldStyle": "font-size:1.9em;background-color:green;color:white;",
                    "$diagnoses": [{
                        message: "test info",
                        severity: "info"
                    }]
                }
            }
        }]
    }
}, {
    "$title": "reorder",
    "$id": "reorder",
    delta: {
        $isPartialDelta: true,
        "xarray": []
    }
}, {
    "$title": "delete record2",
    delta: {
        $isPartialDelta: true,
        "xarray": [{
            "$uuid": "record2",
            "$index": 2,
            "$isDeleted": true
        }]
    }
}, {
    "$title": "delete record3 record6 record7",
    delta: {
        $isPartialDelta: true,
        "xarray": [{
            "$uuid": "record3",
            "$index": 3,
            "$isDeleted": true
        }, {
            "$uuid": "record6",
            "$index": 6,
            "$isDeleted": true
        }, {
            "$uuid": "record7",
            "$index": 7,
            "$isDeleted": true
        }]
    }
}, {
    "$title": "single diagnose",
    delta: {
        $isPartialDelta: true,
        "xarray": [{
            "$uuid": "record0",
            "$index": 0,
            "$properties": {
                "xstring": {
                    "$diagnoses": [{
                        "severity": "error",
                        "message": "Error test  blalfoz"
                    }]
                }
            }
        }]
    }
}, {
    "$title": "multi diagnoses",
    delta: {
        $isPartialDelta: true,
        "$properties": {
            "xarray": {
                "$diagnoses": [{
                    "severity": "error",
                    "message": "Error test  blalfoz"
                }]
            }
        },
        "xarray": [{
            "$uuid": "record0",
            "$index": 0,
            "$properties": {
                "xstring": {
                    "$diagnoses": [{
                        "severity": "error",
                        "message": "Error test  blalfoz"
                    }]
                },
                "xchoice": {
                    "$diagnoses": [{
                        "severity": "info",
                        "message": "Info test"
                    }]
                },
                "xdate": {
                    "$diagnoses": [{
                        "severity": "warning",
                        "message": "Warning test"
                    }]
                },
                "xboolean": {
                    "$diagnoses": [{
                        "severity": "error",
                        "message": "Error test  blalfoz"
                    }]
                },
                "xdecimal": {
                    "$diagnoses": [{
                        "severity": "error",
                        "message": "Error test  blalfoz"
                    }]
                }
            }
        }, {
            "$uuid": "record2",
            "$index": 2,
            "$properties": {
                "xstring": {
                    "$diagnoses": [{
                        "severity": "error",
                        "message": "Error test  blalfoz"
                    }]
                },
                "xchoice": {
                    "$diagnoses": [{
                        "severity": "warning",
                        "message": "Warning test"
                    }]
                }
            }
        }, {
            "$uuid": "record4",
            "$index": 4,
            "$properties": {
                "xstring": {
                    "$diagnoses": [{
                        "severity": "warning",
                        "message": "Warning test"
                    }]
                }
            }
        }, {
            "$uuid": "record6",
            "$index": 6,
            "$properties": {
                "xstring": {
                    "$diagnoses": [{
                        "severity": "error",
                        "message": "Error test  blalfoz"
                    }]
                },
                "xdecimal": {
                    "$diagnoses": [{
                        "severity": "error",
                        "message": "Error test  blalfoz"
                    }]
                }
            }
        }]
    }
}, {
    "$title": "$links = null",
    delta: {
        $isPartialDelta: true,
        "$properties": {
            "xarray": {
                "$links": null
            }
        },
        "xarray": [{
            "$uuid": "record1",
			"$index":1,
            "$properties": {
                "xstring": {
                    "$links": null
                },
                "xchoice": {
                    "$links": null
                },
                "xdate": {
                    "$links": null
                },
                "xboolean": {
                    "$links": null
                },
                "xdecimal": {
                    "$links": null
                }
            },
            "$links": null
        }, {
            "$uuid": "record3",
			"$index":3,
            "$properties": {
                "xstring": {
                    "$links": null
                },
                "xchoice": {
                    "$links": null
                },
                "xdate": {
                    "$links": null
                },
                "xboolean": {
                    "$links": null
                },
                "xdecimal": {
                    "$links": null
                }
            },
            "$links": null
        }, {
            "$uuid": "record5",
			"$index":5,
            "$properties": {
                "xstring": {
                    "$links": null
                },
                "xchoice": {
                    "$links": null
                },
                "xdate": {
                    "$links": null
                },
                "xboolean": {
                    "$links": null
                },
                "xdecimal": {
                    "$links": null
                }
            },
            "$links": null
        }, {
            "$uuid": "record7",
			"$index":7,
            "$properties": {
                "xstring": {
                    "$links": null
                },
                "xchoice": {
                    "$links": null
                },
                "xdate": {
                    "$links": null
                },
                "xboolean": {
                    "$links": null
                },
                "xdecimal": {
                    "$links": null
                }
            },
            "xdecimal": 318178,
            "$links": null
        }]
    }
}, {
    "$title": "$links = $lazyload",
    delta: {
        $isPartialDelta: true,
        "$properties": {
            "xarray": {
                "$links": {
                    "$lazyload": {}
                }
            }
        },
        "xarray": [{
            "$uuid": "record1",
			"$index":1,
            "$properties": {
                "xstring": {
                    "$links": {
                        "$lazyload": {}
                    }
                },
                "xchoice": {
                    "$links": {
                        "$lazyload": {}
                    }
                },
                "xdate": {
                    "$links": {
                        "$lazyload": {}
                    }
                },
                "xboolean": {
                    "$links": {
                        "$lazyload": {}
                    }
                },
                "xdecimal": {
                    "$links": {
                        "$lazyload": {}
                    }
                }
            },
            "$links": {
                "$lazyload": {}
            }
        }, {
            "$uuid": "record3",
			"$index":3,
            "$properties": {
                "xstring": {
                    "$links": {
                        "$lazyload": {}
                    }
                },
                "xchoice": {
                    "$links": {
                        "$lazyload": {}
                    }
                },
                "xdate": {
                    "$links": {
                        "$lazyload": {}
                    }
                },
                "xboolean": {
                    "$links": {
                        "$lazyload": {}
                    }
                },
                "xdecimal": {
                    "$links": {
                        "$lazyload": {}
                    }
                }
            },
            "$links": {
                "$lazyload": {}
            }
        }, {
            "$uuid": "record5",
			"$index":5,
            "$properties": {
                "xstring": {
                    "$links": {
                        "$lazyload": {}
                    }
                },
                "xchoice": {
                    "$links": {
                        "$lazyload": {}
                    }
                },
                "xdate": {
                    "$links": {
                        "$lazyload": {}
                    }
                },
                "xboolean": {
                    "$links": {
                        "$lazyload": {}
                    }
                },
                "xdecimal": {
                    "$links": {
                        "$lazyload": {}
                    }
                }
            },
            "$links": {
                "$lazyload": {}
            }
        }, {
            "$uuid": "record7",
			"$index":7,
            "$properties": {
                "xstring": {
                    "$links": {
                        "$lazyload": {}
                    }
                },
                "xchoice": {
                    "$links": {
                        "$lazyload": {}
                    }
                },
                "xdate": {
                    "$links": {
                        "$lazyload": {}
                    }
                },
                "xboolean": {
                    "$links": {
                        "$lazyload": {}
                    }
                },
                "xdecimal": {
                    "$links": {
                        "$lazyload": {}
                    }
                }
            },
            "$links": {
                "$lazyload": {}
            }
        }]
    }
}, {
    "$title": "$links ans $actions",
    delta: {
        $isPartialDelta: true,
        "$properties": {
            "xarray": {
                "$actions": {
                    "$save": {
                        "$title": "Save"
                    }
                },
                "$links": {
                    "google": {
                        "$title": "Google",
                        "$url": "http://www.google.fr/"
                    },
                    "uioverview": {
                        "$title": "UI Overview",
                        "$url": "?representation=s-uitest-fields.$test"
                    },
                    "$details": {
                        "$type": "application/json;vnd.sage=syracuse",
                        "$title": "Detail",
                        "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
                    },
                    "$linksFormParameters": {
                        "$title": "Parameters",
                        "$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
                        "$parameters": {
                            "param1": "xstring",
                            "param2": "value"
                        }
                    }
                }
            }
        },
        "xarray": [{
            "$uuid": "record1",
			"$index":1,			
            "$properties": {
                "xstring": {
                    "$actions": {
                        "$save": {
                            "$title": "Save"
                        }
                    },
                    "$links": {
                        "google": {
                            "$title": "Google",
                            "$url": "http://www.google.fr/"
                        },
                        "uioverview": {
                            "$title": "UI Overview",
                            "$url": "?representation=s-uitest-fields.$test"
                        },
                        "$details": {
                            "$type": "application/json;vnd.sage=syracuse",
                            "$title": "Detail",
                            "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
                        },
                        "$linksFormParameters": {
                            "$title": "Parameters",
                            "$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
                            "$parameters": {
                                "param1": "xstring",
                                "param2": "value"
                            }
                        }
                    }
                },
                "xchoice": {
                    "$actions": {
                        "$save": {
                            "$title": "Save"
                        }
                    },
                    "$links": {
                        "google": {
                            "$title": "Google",
                            "$url": "http://www.google.fr/"
                        },
                        "uioverview": {
                            "$title": "UI Overview",
                            "$url": "?representation=s-uitest-fields.$test"
                        },
                        "$details": {
                            "$type": "application/json;vnd.sage=syracuse",
                            "$title": "Detail",
                            "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
                        },
                        "$linksFormParameters": {
                            "$title": "Parameters",
                            "$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
                            "$parameters": {
                                "param1": "xstring",
                                "param2": "value"
                            }
                        }
                    }
                },
                "xdate": {
                    "$actions": {
                        "$save": {
                            "$title": "Save"
                        }
                    },
                    "$links": {
                        "google": {
                            "$title": "Google",
                            "$url": "http://www.google.fr/"
                        },
                        "uioverview": {
                            "$title": "UI Overview",
                            "$url": "?representation=s-uitest-fields.$test"
                        },
                        "$details": {
                            "$type": "application/json;vnd.sage=syracuse",
                            "$title": "Detail",
                            "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
                        },
                        "$linksFormParameters": {
                            "$title": "Parameters",
                            "$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
                            "$parameters": {
                                "param1": "xstring",
                                "param2": "value"
                            }
                        }
                    }
                },
                "xboolean": {
                    "$actions": {
                        "$save": {
                            "$title": "Save"
                        }
                    },
                    "$links": {
                        "google": {
                            "$title": "Google",
                            "$url": "http://www.google.fr/"
                        },
                        "uioverview": {
                            "$title": "UI Overview",
                            "$url": "?representation=s-uitest-fields.$test"
                        },
                        "$details": {
                            "$type": "application/json;vnd.sage=syracuse",
                            "$title": "Detail",
                            "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
                        },
                        "$linksFormParameters": {
                            "$title": "Parameters",
                            "$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
                            "$parameters": {
                                "param1": "xstring",
                                "param2": "value"
                            }
                        }
                    }
                },
                "xdecimal": {
                    "$actions": {
                        "$save": {
                            "$title": "Save"
                        }
                    },
                    "$links": {
                        "google": {
                            "$title": "Google",
                            "$url": "http://www.google.fr/"
                        },
                        "uioverview": {
                            "$title": "UI Overview",
                            "$url": "?representation=s-uitest-fields.$test"
                        },
                        "$details": {
                            "$type": "application/json;vnd.sage=syracuse",
                            "$title": "Detail",
                            "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
                        },
                        "$linksFormParameters": {
                            "$title": "Parameters",
                            "$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
                            "$parameters": {
                                "param1": "xstring",
                                "param2": "value"
                            }
                        }
                    }
                }
            },
            "$actions": {
                "$save": {
                    "$title": "Save"
                }
            },
            "$links": {
                "google": {
                    "$title": "Google",
                    "$url": "http://www.google.fr/"
                },
                "uioverview": {
                    "$title": "UI Overview",
                    "$url": "?representation=s-uitest-fields.$test"
                },
                "$details": {
                    "$type": "application/json;vnd.sage=syracuse",
                    "$title": "Detail",
                    "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
                },
                "$linksFormParameters": {
                    "$title": "Parameters",
                    "$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
                    "$parameters": {
                        "param1": "xstring",
                        "param2": "value"
                    }
                }
            }
        }, {
            "$uuid": "record3",
			"$index":3,
            "$properties": {
                "xstring": {
                    "$actions": {
                        "$save": {
                            "$title": "Save"
                        }
                    },
                    "$links": {
                        "google": {
                            "$title": "Google",
                            "$url": "http://www.google.fr/"
                        },
                        "uioverview": {
                            "$title": "UI Overview",
                            "$url": "?representation=s-uitest-fields.$test"
                        },
                        "$details": {
                            "$type": "application/json;vnd.sage=syracuse",
                            "$title": "Detail",
                            "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
                        },
                        "$linksFormParameters": {
                            "$title": "Parameters",
                            "$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
                            "$parameters": {
                                "param1": "xstring",
                                "param2": "value"
                            }
                        }
                    }
                },
                "xchoice": {
                    "$actions": {
                        "$save": {
                            "$title": "Save"
                        }
                    },
                    "$links": {
                        "google": {
                            "$title": "Google",
                            "$url": "http://www.google.fr/"
                        },
                        "uioverview": {
                            "$title": "UI Overview",
                            "$url": "?representation=s-uitest-fields.$test"
                        },
                        "$details": {
                            "$type": "application/json;vnd.sage=syracuse",
                            "$title": "Detail",
                            "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
                        },
                        "$linksFormParameters": {
                            "$title": "Parameters",
                            "$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
                            "$parameters": {
                                "param1": "xstring",
                                "param2": "value"
                            }
                        }
                    }
                },
                "xdate": {
                    "$actions": {
                        "$save": {
                            "$title": "Save"
                        }
                    },
                    "$links": {
                        "google": {
                            "$title": "Google",
                            "$url": "http://www.google.fr/"
                        },
                        "uioverview": {
                            "$title": "UI Overview",
                            "$url": "?representation=s-uitest-fields.$test"
                        },
                        "$details": {
                            "$type": "application/json;vnd.sage=syracuse",
                            "$title": "Detail",
                            "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
                        },
                        "$linksFormParameters": {
                            "$title": "Parameters",
                            "$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
                            "$parameters": {
                                "param1": "xstring",
                                "param2": "value"
                            }
                        }
                    }
                },
                "xboolean": {
                    "$actions": {
                        "$save": {
                            "$title": "Save"
                        }
                    },
                    "$links": {
                        "google": {
                            "$title": "Google",
                            "$url": "http://www.google.fr/"
                        },
                        "uioverview": {
                            "$title": "UI Overview",
                            "$url": "?representation=s-uitest-fields.$test"
                        },
                        "$details": {
                            "$type": "application/json;vnd.sage=syracuse",
                            "$title": "Detail",
                            "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
                        },
                        "$linksFormParameters": {
                            "$title": "Parameters",
                            "$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
                            "$parameters": {
                                "param1": "xstring",
                                "param2": "value"
                            }
                        }
                    }
                },
                "xdecimal": {
                    "$actions": {
                        "$save": {
                            "$title": "Save"
                        }
                    },
                    "$links": {
                        "google": {
                            "$title": "Google",
                            "$url": "http://www.google.fr/"
                        },
                        "uioverview": {
                            "$title": "UI Overview",
                            "$url": "?representation=s-uitest-fields.$test"
                        },
                        "$details": {
                            "$type": "application/json;vnd.sage=syracuse",
                            "$title": "Detail",
                            "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
                        },
                        "$linksFormParameters": {
                            "$title": "Parameters",
                            "$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
                            "$parameters": {
                                "param1": "xstring",
                                "param2": "value"
                            }
                        }
                    }
                }
            },
            "$actions": {
                "$save": {
                    "$title": "Save"
                }
            },
            "$links": {
                "google": {
                    "$title": "Google",
                    "$url": "http://www.google.fr/"
                },
                "uioverview": {
                    "$title": "UI Overview",
                    "$url": "?representation=s-uitest-fields.$test"
                },
                "$details": {
                    "$type": "application/json;vnd.sage=syracuse",
                    "$title": "Detail",
                    "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
                },
                "$linksFormParameters": {
                    "$title": "Parameters",
                    "$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
                    "$parameters": {
                        "param1": "xstring",
                        "param2": "value"
                    }
                }
            }
        }, {
            "$uuid": "record5",
			"$index":5,
            "$properties": {
                "xstring": {
                    "$actions": {
                        "$save": {
                            "$title": "Save"
                        }
                    },
                    "$links": {
                        "google": {
                            "$title": "Google",
                            "$url": "http://www.google.fr/"
                        },
                        "uioverview": {
                            "$title": "UI Overview",
                            "$url": "?representation=s-uitest-fields.$test"
                        },
                        "$details": {
                            "$type": "application/json;vnd.sage=syracuse",
                            "$title": "Detail",
                            "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
                        },
                        "$linksFormParameters": {
                            "$title": "Parameters",
                            "$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
                            "$parameters": {
                                "param1": "xstring",
                                "param2": "value"
                            }
                        }
                    }
                },
                "xchoice": {
                    "$actions": {
                        "$save": {
                            "$title": "Save"
                        }
                    },
                    "$links": {
                        "google": {
                            "$title": "Google",
                            "$url": "http://www.google.fr/"
                        },
                        "uioverview": {
                            "$title": "UI Overview",
                            "$url": "?representation=s-uitest-fields.$test"
                        },
                        "$details": {
                            "$type": "application/json;vnd.sage=syracuse",
                            "$title": "Detail",
                            "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
                        },
                        "$linksFormParameters": {
                            "$title": "Parameters",
                            "$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
                            "$parameters": {
                                "param1": "xstring",
                                "param2": "value"
                            }
                        }
                    }
                },
                "xdate": {
                    "$actions": {
                        "$save": {
                            "$title": "Save"
                        }
                    },
                    "$links": {
                        "google": {
                            "$title": "Google",
                            "$url": "http://www.google.fr/"
                        },
                        "uioverview": {
                            "$title": "UI Overview",
                            "$url": "?representation=s-uitest-fields.$test"
                        },
                        "$details": {
                            "$type": "application/json;vnd.sage=syracuse",
                            "$title": "Detail",
                            "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
                        },
                        "$linksFormParameters": {
                            "$title": "Parameters",
                            "$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
                            "$parameters": {
                                "param1": "xstring",
                                "param2": "value"
                            }
                        }
                    }
                },
                "xboolean": {
                    "$actions": {
                        "$save": {
                            "$title": "Save"
                        }
                    },
                    "$links": {
                        "google": {
                            "$title": "Google",
                            "$url": "http://www.google.fr/"
                        },
                        "uioverview": {
                            "$title": "UI Overview",
                            "$url": "?representation=s-uitest-fields.$test"
                        },
                        "$details": {
                            "$type": "application/json;vnd.sage=syracuse",
                            "$title": "Detail",
                            "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
                        },
                        "$linksFormParameters": {
                            "$title": "Parameters",
                            "$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
                            "$parameters": {
                                "param1": "xstring",
                                "param2": "value"
                            }
                        }
                    }
                },
                "xdecimal": {
                    "$actions": {
                        "$save": {
                            "$title": "Save"
                        }
                    },
                    "$links": {
                        "google": {
                            "$title": "Google",
                            "$url": "http://www.google.fr/"
                        },
                        "uioverview": {
                            "$title": "UI Overview",
                            "$url": "?representation=s-uitest-fields.$test"
                        },
                        "$details": {
                            "$type": "application/json;vnd.sage=syracuse",
                            "$title": "Detail",
                            "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
                        },
                        "$linksFormParameters": {
                            "$title": "Parameters",
                            "$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
                            "$parameters": {
                                "param1": "xstring",
                                "param2": "value"
                            }
                        }
                    }
                }
            },
            "$actions": {
                "$save": {
                    "$title": "Save"
                }
            },
            "$links": {
                "google": {
                    "$title": "Google",
                    "$url": "http://www.google.fr/"
                },
                "uioverview": {
                    "$title": "UI Overview",
                    "$url": "?representation=s-uitest-fields.$test"
                },
                "$details": {
                    "$type": "application/json;vnd.sage=syracuse",
                    "$title": "Detail",
                    "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
                },
                "$linksFormParameters": {
                    "$title": "Parameters",
                    "$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
                    "$parameters": {
                        "param1": "xstring",
                        "param2": "value"
                    }
                }
            }
        }, {
            "$uuid": "record7",
			"$index":7,
            "$properties": {
                "xstring": {
                    "$actions": {
                        "$save": {
                            "$title": "Save"
                        }
                    },
                    "$links": {
                        "google": {
                            "$title": "Google",
                            "$url": "http://www.google.fr/"
                        },
                        "uioverview": {
                            "$title": "UI Overview",
                            "$url": "?representation=s-uitest-fields.$test"
                        },
                        "$details": {
                            "$type": "application/json;vnd.sage=syracuse",
                            "$title": "Detail",
                            "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
                        },
                        "$linksFormParameters": {
                            "$title": "Parameters",
                            "$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
                            "$parameters": {
                                "param1": "xstring",
                                "param2": "value"
                            }
                        }
                    }
                },
                "xchoice": {
                    "$actions": {
                        "$save": {
                            "$title": "Save"
                        }
                    },
                    "$links": {
                        "google": {
                            "$title": "Google",
                            "$url": "http://www.google.fr/"
                        },
                        "uioverview": {
                            "$title": "UI Overview",
                            "$url": "?representation=s-uitest-fields.$test"
                        },
                        "$details": {
                            "$type": "application/json;vnd.sage=syracuse",
                            "$title": "Detail",
                            "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
                        },
                        "$linksFormParameters": {
                            "$title": "Parameters",
                            "$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
                            "$parameters": {
                                "param1": "xstring",
                                "param2": "value"
                            }
                        }
                    }
                },
                "xdate": {
                    "$actions": {
                        "$save": {
                            "$title": "Save"
                        }
                    },
                    "$links": {
                        "google": {
                            "$title": "Google",
                            "$url": "http://www.google.fr/"
                        },
                        "uioverview": {
                            "$title": "UI Overview",
                            "$url": "?representation=s-uitest-fields.$test"
                        },
                        "$details": {
                            "$type": "application/json;vnd.sage=syracuse",
                            "$title": "Detail",
                            "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
                        },
                        "$linksFormParameters": {
                            "$title": "Parameters",
                            "$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
                            "$parameters": {
                                "param1": "xstring",
                                "param2": "value"
                            }
                        }
                    }
                },
                "xboolean": {
                    "$actions": {
                        "$save": {
                            "$title": "Save"
                        }
                    },
                    "$links": {
                        "google": {
                            "$title": "Google",
                            "$url": "http://www.google.fr/"
                        },
                        "uioverview": {
                            "$title": "UI Overview",
                            "$url": "?representation=s-uitest-fields.$test"
                        },
                        "$details": {
                            "$type": "application/json;vnd.sage=syracuse",
                            "$title": "Detail",
                            "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
                        },
                        "$linksFormParameters": {
                            "$title": "Parameters",
                            "$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
                            "$parameters": {
                                "param1": "xstring",
                                "param2": "value"
                            }
                        }
                    }
                },
                "xdecimal": {
                    "$actions": {
                        "$save": {
                            "$title": "Save"
                        }
                    },
                    "$links": {
                        "google": {
                            "$title": "Google",
                            "$url": "http://www.google.fr/"
                        },
                        "uioverview": {
                            "$title": "UI Overview",
                            "$url": "?representation=s-uitest-fields.$test"
                        },
                        "$details": {
                            "$type": "application/json;vnd.sage=syracuse",
                            "$title": "Detail",
                            "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
                        },
                        "$linksFormParameters": {
                            "$title": "Parameters",
                            "$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
                            "$parameters": {
                                "param1": "xstring",
                                "param2": "value"
                            }
                        }
                    }
                }
            },
            "$links": {
                "google": {
                    "$title": "Google",
                    "$url": "http://www.google.fr/"
                },
                "uioverview": {
                    "$title": "UI Overview",
                    "$url": "?representation=s-uitest-fields.$test"
                },
                "$details": {
                    "$type": "application/json;vnd.sage=syracuse",
                    "$title": "Detail",
                    "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
                },
                "$linksFormParameters": {
                    "$title": "Parameters",
                    "$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}",
                    "$parameters": {
                        "param1": "xstring",
                        "param2": "value"
                    }
                }
            },
            "$actions": {
                "$save": {
                    "$title": "Save"
                }
            }
        }]
    }
}];
