// syracuse-ui/lib/desktop/field/chart/flashChart.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Article = require("syracuse-ui/lib/common/article/article").Article;
var flashWrapper = require("./flashWrapper").flashWrapper;


function RecordArticle(){
}

exports.RecordArticle = helpers.defineClass(RecordArticle, Article, {
    drawBox: function(){
        document.itemFactory.load($("<div style='display:none'/>").appendTo(this.boxParent.$$container), {
            $category: "links"
        }, this);
    },
});


function TabularChart(){
}

exports.TabularChart = helpers.defineClass(TabularChart, Article, {
    _executeRecordMenu: function(context){
        if (!this.recordArticle) {
            this.recordArticle = new RecordArticle();
            this.recordArticle.$prototype = this.$prototype.$item;
            document.itemFactory.initializeItem(this.recordArticle, {}, this);
            this.recordArticle.loadBox(context.$item);
        }
        else {
            this.recordArticle.applyChange(context.$item);
        }
        this.recordArticle.menusFacade.items[Object.keys(context["$links" || "$actions"])[0]][0].click();
    },
    fxgetinitialdata: function(callback, myData){
        callback(null, null, this.$prototype);
    },
    fxgetctxmenu: function(callback, myData, line){
        callback(null, null);
    },
    fxexeclink: function(callback, myData, context){
        this._executeRecordMenu(context);
    },
    fxexecdrill: function(callback, myData, drillParams){
        this.getPage().fetch({
            jsonParams: {
                $axes: drillParams.$axes,
                $slicer: drillParams.$slicer
            }
        });
    },
    fxnamedsvccall: function(callback, myData, svcParams){
        alert("Hi, i'm \"" + myData.id + "\" and y should process named service \"" + svcParams.svcId + "\"");
        callback(null, null);
    },
    drawBox: function(){
        var id = this.id + "_flash";
        this.wrapper = new flashWrapper("graph", {
            cont: this.$$container,
            id: id,
            swf: "/syracuse-ui/addins/sageFlash/swf/FLASH",
            base: ".",
            bgcolor: "#FFFFFF",
            width: "100%",
            height: "100%"
        }, {
            instance: this,
            id: id
        }, {
            dataBinding: this.$item.$bind
        });
        this.$$item = this.wrapper.$$elt;
        this.getArticleParent().bind(this, this.$item.$bind);
    },
    setDataBind: function(value, record, metaData){
        if (value !== undefined) {
            if (value && this.wrapper) {
                this.wrapper.setData(value, this.$prototype, metaData);
            }
        }
    }
});
