// syracuse-ui/lib/desktop/field/chart/highChartsConfig.js
"use strict";
exports.defaultOptions = {
	colors: ['#a9ca22', '#d6edba', '#79b251', '#80699B', '#f2af00', '#dbceac', '#e98300', '#c55e9b', '#c99272','#00a1de','#a0d6d2','#00338d' ],
	title: {
		align: 'center',
		// floating: false,
		// margin: 15,
		// x: 0,
		// verticalAlign: 'top',
		y: 15,
		style: {
			color: '#cccccc',
			fontSize: '16px'
		}
	},
	chart: {
		borderColor: '#4572A7',
		//borderWidth: 6,
		borderRadius: 10,
		spacingTop: 30,
		spacingRight: 10,
		spacingBottom: 15,
		spacingLeft: 10,
		style: {
			fontFamily: '"Segoe UI",Tahoma,arial,sans-serif',
			// default font
			fontSize: '11px',
		},
		backgroundColor: '#333',
		//plotBackgroundColor: null,
		plotBorderColor: '#C0C0C0',
		//plotBorderWidth: 0,
		//plotShadow: false,
		//zoomType: ''
	},

	subtitle: {
		align: 'center',
		// floating: false
		// x: 0,
		// verticalAlign: 'top',
		y: 30,
		style: {
			color: '#ccc',
			fontSize: '20px',
		}
	},
	xAxis:{
	
			lineColor:'#ff0000'
	},

	plotOptions: {
		area: {
			fillOpacity: '0.5',
		},
		line: { // base series options
			animation: {
				duration: 1000
			},
			//connectNulls: false,
			//cursor: 'default',
			//clip: true,
			//dashStyle: null,
			//enableMouseTracking: true,
			//legendIndex: 0,
			lineWidth: 2,
			shadow: true,
			// stacking: null,
			marker: {
				//symbol: null,
				lineWidth: 0,
				radius: 4,
				lineColor: '#FFFFFF',
				//fillColor: null,
				states: { // states for a single point
					hover: {
						//radius: base + 2
					},
					select: {
						fillColor: '#FFFFFF',
						lineColor: '#000000',
						lineWidth: 2
					}
				}
			}
		}
	},
	labels: {
		style: {
			//font: defaultFont,
			position: 'absolute',
			color: '#3E576F',
		}
	},
	legend: {
		borderWidth: 1,
		borderColor: 'rgba(250,250,250,0.05)',
		borderRadius: 5,
		// margin: 10,
		// reversed: false,
		shadow: true,
		backgroundColor: 'rgba(0,0,0,0.15)',
		style: {
			padding: '5px'
		},
		itemStyle: {
			cursor: 'pointer',
			color: 'rgba(250,250,250,0.6)',
		    fontFamily:'"georgia"',
			fontStyle: 'italic',
			fontSize: '11px',
			lineHeight: '14px',
		},
		itemHoverStyle: {
			//cursor: 'pointer', removed as of #601
			color: 'rgba(250,250,250,0.8)',
		},
		itemHiddenStyle: {
			color: '#555'
		},
		itemCheckboxStyle: {
			position: 'absolute',
			width: '13px',
			// for IE precision
			height: '13px'
		},
		// itemWidth: undefined,
		symbolWidth: 16,
		symbolPadding: 5,
		verticalAlign: 'bottom',
		// width: undefined,
	},

	loading: {
		// hideDuration: 100,
		labelStyle: {
			fontWeight: 'bold',
			position: 'relative',
			top: '1em'
		},
		// showDuration: 0,
		style: {
			position: 'absolute',
			backgroundColor: 'white',
			opacity: 0.5,
			textAlign: 'center'
		}
	},

	tooltip: {
		//crosshairs: null,
		backgroundColor: 'rgba(50, 50,50, .9)',
		borderWidth: 2,
		borderRadius: 5,
		shadow: true,
		style: {
			color: '#ccc',
			fontFamily:'"georgia"',
			fontStyle: 'italic',
			fontSize: '12px',
			padding: '5px',
			whiteSpace: 'nowrap',
		}
	}
};
