// syracuse-ui/lib/page/utility/designerViewChoice.js
"use strict";

function _addView($view) {

}

exports.onDesignViewClick = function(page, event, btn) {

};
exports.onItemInOut = function(onEnter, event, target) {
	syra_site.dom.toggleClass(target, "s-record-enter", onEnter);
	syra_menus.onTipItemInOut(onEnter, event, target);
};

exports.toggle = function(page) {
	if (!page.designViewsPopup) {
		var slot = document.createElement("aside");
		slot.className = "s-designer-views";
		var header = document.createElement("header");
		header.className = "s-designer-views-header";
		slot.appendChild(header);
		var body = document.createElement("div");
		body.className = "s-designer-views-body";
		slot.appendChild(body);
		var factory;
		var more = document.createElement("div");
		more.className = "s-designer-views-more";
		body.appendChild(more);
		for (var ii = 0, jj = page.$views.length; ii < jj; ii++) {
			var $view = page.$views[ii];
			if ($view.$isFactory) {
				factory = syra_menus.addTextButton($view.$title, "s-designer-view-factory", "onDesignViewClick");
				factory.syraViewIndex = ii;
				body.insertBefore(factory, body.firstChild);
			} else {
				if ($view.$selected) {
					var view = document.createElement("div");
					view.className = "s-designer-view-flag-selected";
					view.textContent = view.title = $view.$title;
					view.syraTip = 1;
					var div = document.createElement("div");
					div.className = "s-designer-view-flag-selected";
					view.className = "s-designer-view-selected";
					view.appendChild(div);
					body.insertBefore(view, more);
				} else {
					var view = syra_menus.addTextButton($view.$title, "s-designer-view", "onDesignViewClick");
					view.syrainout = page.id;
					view.syraViewIndex = ii;

					var btn = syra_menus.addIconButton("edit", "s-designer-view-btn", "onDesignViewClick", null, "$edit");
					view.appendChild(btn);
					btn = syra_menus.addIconButton("delete", "s-designer-view-btn", "onDesignViewClick", null, "$delete");
					view.appendChild(btn);
					more.appendChild(view);
				}
			}

		}
		page.designViewsPopup = syra_site.dialogManager.openPopup(page, {
			content: {},
			slot: slot,
			position: {
				my: "right top",
				at: "right bottom",
				of: $(page.designerViewOpener)
			},
			onClose: function() {
				syra_site.dom.empty(slot);
				page.designViewsPopup = null;
			}
		});
	}
};
