// syracuse-ui/lib/page/utility/importHelper.js
"use strict";


function import_navigationMenu($article, $prototype, page) {
	page.createDefaultArticle($article, $prototype);
	var $items = $article.$layout.$items;
	if ($items.length == 1) {
		var $sections = $items[0].$items;
		$article.$layout.$items = [];
		$article.$layout.$items.push({
			$layoutType: "stack",
			$items: []
		}, {
			$layoutType: "stack",
			$items: []
		});
		var half = Math.ceil($sections.length / 2);
		for (var ii = 0, jj = $sections.length; ii < jj; ii++) {
			$article.$layout.$items[ii < half ? 0 : 1].$items.push($sections[ii]);
		}
		$article.$layout.$layoutType = "row";
	}
	return $article;
}

function _importVignettes($items, $vignettes, $prototype, page) {
	var $row;
	for (var ii = 0, jj = $vignettes.length; ii < jj; ii++) {
		var $field = $prototype.$properties[$vignettes[ii]];
		if ($field) { // && !$field.$isExcluded && !$field.$isTOC) {
			if (!$row) {
				$row = {
					$layoutType: "row",
					$items: []
				};
				$items.push($row);
			}
			$row.$items.push(page.addDefaultField($vignettes[ii], $field));
			if ($row.$items.length == 2) {
				$row = null;
			}
		} else {
			//debugger;
		}
	}
}

function import_mainDashboard($article, $prototype, page) {
	var $items = [];
	if ($article.$contentHint) {
		var $tabs = $article.$contentHint.$tabs;
		for (var ii = 0, jj = $tabs.length; ii < jj; ii++) {
			var $tab = $tabs[ii];
			var $sectionItems = [];
			if ($tab.$queries) {
				_importVignettes($sectionItems, $tab.$queries, $prototype, page);
			}
			if ($tab.$processes) {
				_importVignettes($sectionItems, $tab.$processes, $prototype, page);
			}
			$items.push({
				$category: "section",
				$title: $tab.title,
				$layout: {
					$layoutType: "stack",
					$items: $sectionItems
				}
			});
		}
		delete $article.$contentHint;
		$article.$layout = {
			$layoutType: "tabs",
			$items: $items
		};
	}
	return $article;
}

exports.importArticle = function($article, $prototype, page) {
	if ($article.$importHint) {
		switch ($article.$importHint) {
			case "navigationMenu":
				import_navigationMenu($article, $prototype, page);
				break;
			case "mainDashboard":
				import_mainDashboard($article, $prototype, page);
				break;
		}
		delete $article.$importHint;
	}
	return $article;
};
