// syracuse-ui/lib/page/utility/menuBar.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var BarSplitter = require("syracuse-ui/lib/page/utility/barSplitter").BarSplitter;

function MenuBar() {}

exports.MenuBar = helpers.defineClass(MenuBar, BarSplitter, {
	load: function(page, layoutSlot, barSlot, barBody) {
		this.options = {
			resizeDirection: {
				left: true
			},
			$skins: {
				bar: page.$skin + "-menubar",
				openerPicker: page.$skin + "-menubar-opener",
				resizeBar: page.$skin + "-menubar-resizer"
			}
		};
		page.initializeNewItem(this);
		var preferences = document.site.ensurePreferences() || {};
		if (!preferences.menuBar) {
			preferences.menuBar = {
				isDocked: true,
				isCollapsed: false
			};
		};
		this.preferences = preferences.menuBar;

		this.barSlot = document.createElement("div");
		this.barSlot.className = page.$skin + "-menubar-slot";
		this.barSlot.style.display = "none";

		this.barBody = document.createElement("div");
		this.barBody.className = page.$skin + "-menubar-body";

		if (page.$pageCategory == "vignette" || page.$pageCategory == "dashboard") {
			this.page._item.insertBefore(this.barSlot, this.page._item.firstChild);
			if (page.$pageCategory == "vignette") {
				this.preferences = {}; //temp for preference management
			}
		} else {
			this.page._item.appendChild(this.barSlot);
		}
		BarSplitter.prototype.load.call(this, page);
	},
	onBarBodyAdded: function() {
		this.page.appendArticleMenus(this.barBody);
	},
	onMenuDataFilled: function(menuItem) {
		if (!this.isSlotVisible && !this.isHidden) {
			if (menuItem.boxParent.menuGroupRoot) {
				if (menuItem.boxParent.menuGroupRoot.$item.$isMenuPage) {
					this.isSlotVisible = true;
					if (!this.page.designer) {
						this.barSlot.style.display = "";
						if (this.page.layoutSlot.style.display != "none") {
							document.site.resize();
						}
					}
				}
			}
		}
	}
});
