// syracuse-ui/lib/page/utility/userPreferences.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function UserPreferences() {}

exports.UserPreferences = helpers.defineClass(UserPreferences, null, {
	load: function(page, $link, data) {
		this.page = page;
		this.$link = $link;
		this.data = data || {};
	},
	getForKey: function(key, defaultValue) {
		if (!this.data[key]) {
			this.data[key] = defaultValue;
		}
		return this.data[key];
	},
	getForField: function($bind) {
		return (!this.page.isVignettePage && this.data.$fields && this.data.$fields[$bind]) || {};
	},
	saveForField: function($bind, $fieldUserPreferences) {
		if (!this.page.isVignettePage) {
			(this.data.$fields = this.data.$fields || {})[$bind] = $fieldUserPreferences;
		}
	},
	applyToMenuGroup: function(menuGroup) {
		if (!menuGroup.$item.$isPopupContent) {
			if (this.data && this.data.$menuBarGroup) {
				var $opened = this.data.$menuBarGroup[menuGroup.id];
				if ($opened != undefined && menuGroup.$opened != $opened) {
					menuGroup.openBox(menuGroup.$opened = $opened);
				}
			}
		}
	},
	save: function() {
		var self = this;
		if (self.$link) {
			var sendBag = {
				$etag: 1,
				$url: syra_site.urlMaker.formatMenuUrl(self.page, self.$link),
				content: self.data
			};
			syra_controller.sendRequest(self.page, {
				$location: {
					$url: sendBag.$url
				},
				data: sendBag,
				method: "PUT",
				$etag: sendBag.$etag
			}, function(data, response, requestUrl) {
				if (!self.page.disposed) {
					//debugger;

				}
			});
		}
	},
	dispose: function() {
		this.page = this.$link = this.data = null;
	}
});
