// syracuse-ui/lib/field/vignette/boWebIVignetteField.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var VignetteField = require("syracuse-ui/lib/field/vignette/vignetteField").VignetteField;

var TUNNEL_DESC = {
	"spgm": "INTRUTILA:BOLINK",
	"params": [{ // FONCTION
		"dim": 1,
		"typ": "STRU",
		"siz": 8
	}, { // VARIA
		"dim": 1,
		"typ": "STRU",
		"siz": 8
	}, { // CLE
		"dim": 1,
		"typ": "STRU",
		"siz": 250
	}]
};

function BOWebIVignetteField() {}

exports.BOWebIVignetteField = helpers.defineClass(BOWebIVignetteField, VignetteField, {
	ensureDefaultTitle: function() {

	},
	setDataValue: function(value, metaData) {
		syra_fields.applyMetaData(this, metaData);
		if (value !== undefined) {
			this.$contentUrl = this.$item.$url = value;
			this._showUrl();
		}
	},
	applyDesignMetaData: function(metaData, designing) {
		VignetteField.prototype.applyDesignMetaData.call(this, metaData, designing);
		if (metaData.$url !== undefined) {
			this.$contentUrl = this.$item.$url = metaData.$url;
			this._showUrl();
		}
	},
	_showUrl: function() {
		var self = this;
		if (self.$contentUrl) {
			if (!self.boWindow) {
				window.x3boSendAckCallTunnel = function(func, value, key) {
					setTimeout(function() {
						syra_fusion.onBlockExRpc(self, TUNNEL_DESC, [func, value, key]);
						self.boWindow && self.boWindow.blur();
						window.focus();
					}, 100);
				};
				self.boWindow = window.open(self.$contentUrl, "_blank");
			}
		} else {
			syra_dom.empty(self.body);
			self.boWindow = null;
		}
	},
	renderLayoutContent: function() {
		VignetteField.prototype.renderLayoutContent.call(this);
		this.$contentUrl = this.$field.$location && this.$field.$location.$url;
		this._showUrl();
	},
	setFocus: function() {
		return true;
	},
	isDirty: function(value, record, metaData, $bind) {
		return false;
	},
	dispose: function() {
		window.sendAckCallTunnel = this.boWindow = null;
		VignetteField.prototype.dispose.call(this);
	}
});
