// syracuse-ui/lib/field/vignette/htmlVignetteField.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var VignetteField = require("syracuse-ui/lib/field/vignette/vignetteField").VignetteField;


function HTMLVignetteField(){
}

exports.HTMLVignetteField = helpers.defineClass(HTMLVignetteField, VignetteField, {
    applyDesignMetaData: function(metaData, onAuthoring){
        VignetteField.prototype.applyDesignMetaData.call(this, metaData, onAuthoring);
        var iframe = this._iframe ? this._iframe : null;
        if (metaData.$url) {
            this.$item.$url = metaData.$url;
            if (this._iframe) {
                this._iframe.setAttribute("src", this.$item.$url);
            }
        }
        if (metaData.$width) {
            this.$item.$width = metaData.$width;
            if (this._iframe) {
                this._iframe.style.width = this.$item.$width;
            }
        }
        if (metaData.$height) {
            this.$item.$height = metaData.$height;
            if (this._iframe) {
                this._iframe.style.height = this.$item.$height;
            }
        }
    },
    onWindowResize: function(){
        var body = document.site.$$body;
        var maxHeight = document.site.$$body.height();
        this._iframe.style.height = maxHeight + "px";
        this._iframe.style.height = (maxHeight - (body[0].scrollHeight - maxHeight)) + "px";
    },
    doPicker: function(pickerId){
        VignetteField.prototype.doPicker.call(this, pickerId);
        switch (pickerId) {
            case "maximize":
                this.onWindowResize();
                document.site.addResizeListener(this);
                break;
            case "minimize":
                this._iframe.style.height = this.$item.$height || "100%";
                document.site.removeResizeListener(this);
                break;
        }
    },
    renderLayoutContent: function(){
        VignetteField.prototype.renderLayoutContent.call(this);
        var url = this.$item.$url || this.$field.$location.$url;
        var title = this.domTitle.textContent;
        this.isTitleUrlLink = true;
        document.site.toggleClass(this.domTitle, "s-title-link", true);
        this.domTitle.setAttribute("href", url);
        this.domTitle.setAttribute("target", "_blank");
        this.domTitle.setAttribute("title", this.$item.$title || this.$field.$location.$title);
        this._iframe = document.createElement("iframe");
        this._iframe.setAttribute("src", url);
        this._iframe.style.width = this.$item.$width || "100%";
        this._iframe.style.height = this.$item.$height || "100%";
        this.$$body.empty().append(this._iframe);
    },
    dispose: function(){
        document.site.removeResizeListener(this);
        VignetteField.prototype.dispose.call(this);
    }
});
