// syracuse-ui/lib/field/vignette/menusVignetteBuilder.js
"use strict";
exports.load = function(field){
    if (field.$field.$location) {
        document.controller.loadRepresentation(null, field.$field.$location.$url, function($itemPage){
            $itemPage.$$container = field.$$body;
            $itemPage.$category = "vignette";
            $itemPage.$skin = $itemPage.$skin || "s-vignette-widget-menus";
            field.vignette = document.site.loadNewPage($itemPage);
        });
    }
    else {
        debugger;
        field.vignette = document.site.loadNewPage({
            $$container: field.$$body,
            $representation: {
                $prototype: {
                    $links: field.$field.$links,
                    $actions: field.$field.$actions
                },
                $article: {
                    $category: "vignette",
                    $layout: {
                        $items: [{
                            $skin: "s-vignette-widget-menus",
                            $category: "links",
                            $menus: field.$field.$menus
                        }]
                    }
                }
            },
            $urlParts: {
                $facet: "$menu"
            },
            $category: "vignette"
        });
    }
};
exports.getAuthoringWidget = function(field){
    return null;
};
exports.dispose = function(){
};
