// syracuse-ui/lib/field/vignette/pageVignetteField.js
"use strict";
var _helpers = require('syracuse-core').helpers;
var VignetteField = require("syracuse-ui/lib/field/vignette/vignetteField").VignetteField;

function _loadContentPage(field) {
	if (!field._isContentPageRequested && syra_dom.isVisible(field.domItem)) {
		field._isContentPageRequested = true;
		if (field.$url || !syra_fields.embedded.hasMasterChildRelation(field)) {
			//if masterChild, vignette is lazy loaded
			field.loadContentRepresentation();
		}
	}
}

function PageVignetteField() {}

exports.PageVignetteField = _helpers.defineClass(PageVignetteField, VignetteField, {
	onAfterMaximize: function() {
		this.vignette && this.vignette.resizeItem(true);
	},
	loadContentRepresentation: function() {
		var self = this;
		syra_router.loadRepresentation({
			article: null,
			segments: self.$url || self.$field.$location.$url,
			success: function($itemPage) {
				if (self.body) {
					$itemPage.layoutSlot = self.body;
					$itemPage.$category = "vignette";
					$itemPage.vignetteField = self;
					var $vignette = self.$item.$vignette;
					var altVignetteKey;
					if (self.$url && self.$field.$location.$url != self.$url) {
						var newSeg = syra_url.parse(self.$url);
						var sourceSeg = syra_url.parse(self.$field.$location.$url);
						if (newSeg.params.representation != sourceSeg.params.representation) {
							$vignette = self.$item.$altVignettes && self.$item.$altVignettes[newSeg.params.representation];
							altVignetteKey = newSeg.params.representation;
							if (!$vignette && self.$item.$vignette) {
								$vignette = syra_site.clone(self.$item.$vignette);
							}
						}
					}
					if ($vignette) {
						$itemPage.$representation.$article = $vignette;
						delete $itemPage.$representation.$article.$menus;
					}
					syra_pageBuilder.load($itemPage, function(page) {
						self.vignette = page;
						self.vignette.altVignetteKey = altVignetteKey;
						self._onVignetteRefresh && self.onItemInOut(true);
					});
				}
			}
		});
	},
	designArticle: function(open) {
		if (this.vignette) {
			switch (this.vignette.$facet) {
				case "$query":
				case "$cube":
					var fields = this.vignette.layoutContent.getFields();
					for (var ii = 0, jj = fields.length; ii < jj; ii++) {
						if (fields[ii].isArrayField) {
							var open = !fields[ii].designer;
							fields[ii].designArticle(open);
							break;
						}
					}
					break;
			}
		}
	},
	renderLayoutContent: function($url, onVignetteRefresh) {
		VignetteField.prototype.renderLayoutContent.call(this);
		if ($url) {
			this.$url = $url;
		} else {
			if (!(this.$url && this.$masterBinds)) {
				this.$url = this.$field.$altLocation && this.$field.$altLocation.$url;
				if (this.$url && this.page.ldpRecord) {
					this.$url = syra_expression.parse(this.page.ldpRecord.page, this.$url);
				}
			}
		}
		this._isContentPageRequested = false;
		this._onVignetteRefresh = onVignetteRefresh;
		_loadContentPage(this);
	},
	resizeItem: function(doResize) {
		_loadContentPage(this);
		this.vignette && this.vignette.resizeItem(doResize);
	},
	dispose: function() {
		syra_fields.embedded.disposeMasterChildRelation(this);
		this.vignette && this.vignette.dispose();
		VignetteField.prototype.dispose.call(this);
	}
});
