// syracuse-ui/lib/field/vignette/vignetteField.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var SectionBlock = require("syracuse-ui/lib/article/sectionBlock").SectionBlock;

function VignetteField() {}

exports.VignetteField = helpers.defineClass(VignetteField, SectionBlock, {
	ensureSkin: function() {
		SectionBlock.prototype.ensureSkin.call(this);
		this.$skinSlot = this.$skin;
		this.$designLevel = "field";
	},
	ensureVisibility: function() {
		//specific behavior for vignetteField that are section and field
		return true;
	},
	resizeField: function() {},
	onMenusVignetteClick: function(event, picker) {
		this.vignette && this.vignette.onMenusVignetteClick(event, picker);
	},
	setDataBind: function(value, record, metaData, $bind) {},
	loadBox: function() {
		SectionBlock.prototype.loadBox.call(this);
		if (this.page.registerVignette) {
			this.page.registerVignette(this);
		} else {
			this.articleParent.bind(this, this.$item.$bind);
		}
		this.articleParent.registerAutoSizeFields(this, true);
	},
	dispose: function() {
		this.articleParent.registerAutoSizeFields(this, false);
		this.vignette = this.$field = null;
		SectionBlock.prototype.dispose.call(this);
	},
	refresh: function(onVignetteRefresh) {
		this.renderLayoutContent(null, onVignetteRefresh);
	},
	onDesignOpenerClick: function() {
		var self = this;
		if (self.page.isLandingPage) {
			syra_site.designStore.design(self, !self.designer);
		} else {
			syra_site.urlMaker.formatParameters(self.articleParent, {
				$properties: {
					vignettUrl: {
						$title: "url",
						$type: "text/plain"
					}
				},
				vignettUrl: self.$field.$location.$url
			}, self.articleParent, function(values) {
				if (values.vignettUrl != selfself.$field.$location.$url) {
					self.$item.$location = self.$item.$location || {};
					selfself.$item.$location.$url = values.vignettUrl;
					self.onRefreshClick();
				}
			});
		}
	},
	onDeleteClick: function() {
		syra_site.landingPageMaster.onDeleteVignetteClick(this);
	},
	onRefreshClick: function() {
		this.refresh(true);
	},
	onItemInOut: function(onEnter, event) {

	},
	ensureButtonsVisibility: function() {
		this.onItemInOut();
	},
	drawBox: function() {
		if (syra_site.isTabletDevice) {
			this.btnWidth = 25;
		}
		if (!this.page.isLandingPage && this.$item.$isBoxCollapsable == undefined) {
			this.$item.$isBoxCollapsable = true;
		}
		this.domItem = document.createElement("section");
		this.domItem.className = this.$skin;
		this.domItem.syrainout = this.id;

		this.layoutSlot.appendChild(this.domItem);
		this.setState(this.$item);
		this.body = document.createElement("div");
		this.body.className = this.$skin + "-body";
		this.appendHeader();
		this.ensureLayoutMode();
		this.domItem.appendChild(this.body);
		this.applyDesignMetaData(this.$item, false);
		if (this.page.isDesignModeEnabled) {
			this.showButton("design", true);
		}
		this.showButton("refresh", true);
		this.showButton("maximize", true);
		if (this.page.isDesignModeEnabled) {
			this.showButton("delete", true, null, "delete");
		}
		this.openBox(this.$item.$opened !== false);
	},

	ensureLayoutMode: function() {
		this.ensureSkin();
		this.layoutParent && this.layoutParent.ensureLayoutSlot(this);
	},
	renderLayoutContent: function() {
		if (this.vignette) {
			syra_site.dom.empty(this.body);
			this.vignette.dispose();
			this.vignette = null;
		}
	},
	toggleUIDesign: function(enable, disposingDesigner) {
		return; // temporaly deactivate
		if (!this.page.isLandingPage) {
			if (enable) {
				if (!this.btns.items.design) {
					this.showButton("design", true);
					this.btns.slot.insertBefore(this.btns.items.design, this.btns.slot.firstChild);
				}
			} else {
				if (this.btns.items.design) {
					syra_site.dom.removeChild(this.btns.items.design);
					delete this.btns.items.design;
				}
			}
		}
	},
	enableDesigner: function(enable) {
		this.showButton("design", this.page.isDesignModeEnabled = false);
	}
});
