// syracuse-ui/lib/site/behavior/dragDrop.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
function DragDrop(){
}

exports.DragDrop = helpers.defineClass(DragDrop, null, {
    setDraggable: function(source, $$hotspot){
        var self = this;
        self.id = helpers.uuid.generate();
        self.source = source;
        (self.$$hotSpot = $$hotspot || $(self.source.layoutSlot)).bind("mousedown." + self.id, function(event){
            var offset = $(self.source.layoutSlot).offset();
            self.startPosition = {
                top: offset.top - event.pageY,
                left: offset.left - event.pageX,
            };
            self.start(self.source.layoutSlot);
            return false;
        });
        self.$$hotSpot[0].style.cursor = "move";
    },
    start: function(slot, $$boundary){
        var self = this;
        self.slot = slot;
        
        $$boundary = $$boundary || document.site.$$body;
        self.boundary = $$boundary.offset();
        self.boundary.right = self.boundary.left + $$boundary.width();
        self.boundary.bottom = self.boundary.top + $$boundary.height();
        $(document).bind("mouseup." + self.id, function(){
            self.stop();
            delete document.site.DDAuthoring;
        }).bind("mousemove." + self.id, function(event){
            self._moveDragImage(event);
            event.preventDefault();
        });
    },
    _moveDragImage: function(event, isValidate){
        var position = {
            top: (this.startPosition.top + event.pageY),
            left: (this.startPosition.left + event.pageX)
        };
        this.slot.style.top = position.top + "px";
        this.slot.style.left = position.left + "px";
        if (this.source) {
            this.source.draggedPosition = position;
        }
        
    },
    stop: function(){
        $(document).unbind("." + this.id);
    },
    dispose: function(){
        if (this.$$hotSpot) {
            this.$$hotSpot.unbind("." + this.id);
        }
        delete this.source;
        this.stop();
        
        delete this.slot;
    }
});
