// @sage/syracuse-react/src/navigation/bookmark/bookmarkMenu.ts
"use strict";
import * as Sdata from '../../dataStore/sdata';
import { Url } from '../../route/url';
import * as Endpoint from '../../profile/endpoint';

export class BookmarkMenu {
    private _url: Url;
    private _enabled: boolean;
    private _dataset: Sdata.IBookmark;

    constructor(dataset: Sdata.IBookmark, profileEndpoint: Endpoint.IEndpoint) {
        this._dataset = dataset;
        let url = this._url = new Url(dataset.$url)
        if (url.host) {
            //set relative url
            url.uri = url.uri.slice(url.uri.indexOf(url.host) + url.host.length);
            url.host = null;
        }
        if (!url.endpoint
            || Endpoint.matchContract(Endpoint.administration, url.endpoint)
            || Endpoint.matchContract(Endpoint.search, url.endpoint)
            || Endpoint.matchContract(Endpoint.x3js, url.endpoint)) {
            this._enabled = true;
        }
        else {
            if (Endpoint.matchContract(url.endpoint, profileEndpoint)) {
                url.endpoint = Endpoint.cloneKey(profileEndpoint);
                url.stringify();
                this._enabled = true;
            }
        }
    }
    get dataset(): Sdata.IBookmark {
        return this._dataset;
    }
    get enabled(): boolean {
        return this._enabled;
    }
    get url(): Url {
        return this._url;
    }
    get title(): string {
        return this.dataset.title;
    }
    get description(): string {
        return this.dataset.description;
    }
}
