// @sage/syracuse-react/src/navigation/bookmark/bookmarkHeader.tsx
import * as React from 'react';
import { BookmarkMenu } from './bookmarkMenu';
import { Bookmarks } from './bookmarks';
import { Link } from '../../link/link';

export class BookmarkHeader extends React.Component<{ bookmarks: Bookmarks; onBookmarkClick: (value: BookmarkMenu) => void }, {}> {
    public width: number;

    constructor(props: any) {
        super(props);
        this.onResize = this.onResize.bind(this);
    }

    public getWidth() {
        return document.querySelector('.s_profile_bar_bookmarks_list').clientWidth;
    }

    public onResize() {
        this.width = this.getWidth();
    }

    public componentDidMount() {
        this.width = this.getWidth();
        if (typeof window !== 'undefined') {
            window.addEventListener('resize', this.onResize, false);
        }
    }
    public componentWillUnmount() {
        if (typeof window !== 'undefined') {
            window.removeEventListener('resize', this.onResize);
        }
    }

    public render() {
        return (
            <div className="s_profile_bar_bookmarks">
                <BookmarkHeaderList bookmarks={this.props.bookmarks} onBookmarkClick={this.props.onBookmarkClick} />
                <BookmarkHeaderIcons />
            </div>
        );
    }
}

export class BookmarkHeaderList extends React.Component<{ bookmarks: Bookmarks; onBookmarkClick: (value: BookmarkMenu) => void }, {}> {
    public render() {
        const bookmarksList = (menu: BookmarkMenu, index: number) =>
            <BookmarkHeaderListItem key={index} title={menu.title} menu={menu} onBookmarkClick={this.props.onBookmarkClick} />;
        return (
            <ul className="s_profile_bar_bookmarks_list">
                {this.props.bookmarks.menus.map(bookmarksList)}
            </ul>
        );
    }
}

export class BookmarkHeaderListItem extends React.Component<
    { title: string; menu: BookmarkMenu; onBookmarkClick: (value: BookmarkMenu) => void },
    {}
> {
    public render() {
        return (
            <li>
                <Link
                    className="s_profile_bar_bookmarks_textlink"
                    title={this.props.title}
                    value={this.props.menu}
                    onClick={this.props.onBookmarkClick}
                />
            </li>
        );
    }
}
export class BookmarkHeaderIcons extends React.Component<{}, {}> {
    public render() {
        return (
            <div className="s_profile_bar_bookmarks_icon">
                <Link className="s_profile_bar_bookmarks_textlink" icon="dir_down" description="Display header" onClick={() => {}} />
                <Link className="s_profile_bar_bookmarks_textlink" icon="dir_down" description="Display popup" onClick={() => {}} />
                <Link className="s_profile_bar_bookmarks_textlink" icon="dir_down" description="Display splitter" onClick={() => {}} />
                <Link className="s_profile_bar_bookmarks_textlink" icon="close" description="Close" onClick={() => {}} />
            </div>
        );
    }
}
