// @sage/syracuse-react/src/navigation/bookmark/bookmarkHeader.tsx
import * as React from 'react';
import { BookmarkMenu } from './bookmarkMenu';
import { Bookmarks } from './bookmarks';
import { Link } from '../../link/link';

export interface IBookmarkHeaderProps {
    bookmarks: Bookmarks;
    onBookmarkClick: (value: BookmarkMenu) => void;
}
export interface IBookmarkHeaderState {
    width: number;
}

export class BookmarkHeader extends React.Component<IBookmarkHeaderProps, IBookmarkHeaderState> {
    public static getWidth(): number {
        return document.querySelector('.s_profile_bar_bookmarks_list').clientWidth;
    }

    constructor(props: any) {
        super(props);
        this.onResize = this.onResize.bind(this);
    }

    public onResize() {
        this.state = { width: BookmarkHeader.getWidth() };
    }

    public componentDidMount() {
        this.state = { width: BookmarkHeader.getWidth() };
        if (typeof window !== 'undefined') {
            window.addEventListener('resize', this.onResize, false);
        }
    }
    public componentWillUnmount() {
        if (typeof window !== 'undefined') {
            window.removeEventListener('resize', this.onResize);
        }
    }

    public render() {
        return (
            <div className="s_profile_bar_bookmarks">
                <BookmarkHeaderList bookmarks={this.props.bookmarks} onBookmarkClick={this.props.onBookmarkClick} />
                <BookmarkHeaderIcons />
            </div>
        );
    }
}

export interface IBookmarkHeaderListProps {
    bookmarks: Bookmarks;
    onBookmarkClick: (value: BookmarkMenu) => void;
}

export class BookmarkHeaderList extends React.Component<IBookmarkHeaderListProps, {}> {
    public render() {
        const bookmarksList = (menu: BookmarkMenu, index: number) =>
            <BookmarkHeaderListItem key={index} menu={menu} onBookmarkClick={this.props.onBookmarkClick} />;
        return (
            <ul className="s_profile_bar_bookmarks_list">
                {this.props.bookmarks.menus.map(bookmarksList)}
            </ul>
        );
    }
}

export interface IBookmarkHeaderListItem {
    menu: BookmarkMenu;
    onBookmarkClick: (value: BookmarkMenu) => void;
}

export class BookmarkHeaderListItem extends React.Component<IBookmarkHeaderListItem, {}> {
    public render() {
        return (
            <li>
                <Link
                    className="s_profile_bar_bookmarks_textlink"
                    title={this.props.menu.title}
                    value={this.props.menu}
                    onClick={this.props.onBookmarkClick}
                />
            </li>
        );
    }
}

export class BookmarkHeaderIcons extends React.Component<{}, {}> {
    public render() {
        return (
            <div className="s_profile_bar_bookmarks_icon">
                <Link className="s_profile_bar_bookmarks_textlink" icon="dir_down" description="Display header" onClick={() => {}} />
            </div>
        );
    }
}
