// @sage/syracuse-react/src/navigation/bookmark/bookmarkController.ts
"use strict";
import * as Sdata from '../../dataStore/sdata';
import { Bookmarks } from './bookmarks';
import { Url } from '../../route/url';
import { Profile } from '../../profile/profile';
import { Ajax, AjaxResponse } from '../../ajax/ajax';

let ajax: Ajax;

export function fetch(profile: Profile, done: (  diagnoses?: Sdata.IDiagnose[]) => void) {
    ajax = new Ajax();
    ajax.get({
        url: profile.parse(profile.links.$bookmarks.$url),
        end: (response: AjaxResponse) => {
            if (response.ok) {
                profile.bookmarks = new Bookmarks(response.data, profile.selectedEndpoint);
            }
            done(response.error && response.error.diagnoses);
        }
    });
}

export function toggleMenu(profile: Profile, url: Url, title: string, description: string, done?: () => void) {
    let bookmarks = profile.bookmarks;
    let menu = url.isFusion ? bookmarks.findClassic(url) : bookmarks.findSyracuse(url);
    if (menu) {
        bookmarks.items.splice(bookmarks.menus.indexOf(menu), 1);
    }
    else {
        let dataset: Sdata.IBookmarkItem = {
            menuItem: {
                $url: url.url,
                title: title
            }
        };
        if (description) {
            dataset.menuItem.description = description;
        }
        bookmarks.items.push(dataset);
    }
    save(profile, done);
}

export function save(profile: Profile, done?: () => void) {
    ajax = new Ajax();
    ajax.put({
        lock: true,
        url: profile.bookmarkUrl,
        data: {
            content: profile.bookmarks.dataset.content
        },
        end: (response: AjaxResponse) => {
            if (response.data) {
                profile.bookmarks.refresh(response.data);
                done && done();
            }
        }
    });
}
