// syracuse-tablet/html/js/sdata/entities/test/testGaugeChart4.js
"use strict";

var _data = {
	"salesOp": "75.0000000000",
	"salesValue": "133000.0000000000",
	"salesValueMin": "100000.0000000000",
	"salesValueMax": "150000.0000000000"
};

var _entity = {
	$entityName: "testGaugeChart4",
	$properties: {
		"salesOp": {
			"$title": "sale opportunity",
			"$type": "application/x-decimal",
			"$capabilities": "sort,filter"
		},
		"salesValue": {
			"$title": "sales value",
			"$type": "application/x-decimal",
			"$capabilities": "sort,filter"
		},
		"salesValueMin": {
			"$title": "sales value min",
			"$type": "application/x-decimal",
			"$capabilities": "sort,filter"
		},
		"salesValueMax": {
			"$title": "sales value max",
			"$type": "application/x-decimal",
			"$capabilities": "sort,filter"
		}
	},
	$relations: {},
	$articles: {
		$details: {
			"$layoutType": "stack",
			$items: [{
				"$bind": "salesValue",
				/* GAUGE 4 */
				"$gauge": {
					"$bindMin": "salesValueMin",
					"$bindMax": "salesValueMax",
					"segments": [{
						"valStart": 100000,
						"valEnd": 130000,
						"color": "#DF5353"
					}, {
						"valStart": 130000,
						"valEnd": 140000
					}, {
						"valStart": 140000,
						"valEnd": 150000,
						"color": "#55BF3B"
					}]
				}
			}]
		}
	},
	$services: {
		$read: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve(_data);
			return deferred.promise();
		}
	}
};

exports.entity = _entity;
