// syracuse-tablet/html/js/sdata/entities/test/testGaugeChart3.js
"use strict";

var _data = {
	"salesOp": "75.0000000000",
	"salesValue": "133000.0000000000",
	"salesValueMin": "100000.0000000000",
	"salesValueMax": "150000.0000000000"
};

var _entity = {
	$entityName: "testGaugeChart3",
	$properties: {
		"salesOp": {
			"$title": "sale opportunity",
			"$type": "application/x-decimal",
			"$capabilities": "sort,filter"
		},
		"salesValue": {
			"$title": "sales value",
			"$type": "application/x-decimal",
			"$capabilities": "sort,filter"
		},
		"salesValueMin": {
			"$title": "sales value min",
			"$type": "application/x-decimal",
			"$capabilities": "sort,filter"
		},
		"salesValueMax": {
			"$title": "sales value max",
			"$type": "application/x-decimal",
			"$capabilities": "sort,filter"
		}
	},
	$relations: {},
	$articles: {
		$details: {
			"$layoutType": "stack",
			$items: [{
				"$bind": "salesValue",
				/* GAUGE 3 */
				"$gauge": {
					"$bindMin": "salesValueMin",
					"$bindMax": "salesValueMax",
					"segments": [{
						"propStart": 0,
						"propEnd": 0.75
					}, {
						"propStart": 0.75,
						"propEnd": 0.85
					}, {
						"propStart": 0.85,
						"propEnd": 1
					}]
				}
			}]
		}
	},
	$services: {
		$read: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve(_data);
			return deferred.promise();
		}
	}
};

exports.entity = _entity;
