// syracuse-tablet/html/js/sdata/entities/test/testFormatsReference.js
"use strict";

var _data = {
	SITE: "ASN",
	SITE_REF: {
		$description: "Asnières",
		$title: "Asnières"
	}
};

exports.entity = {
	$entityName: "testFormatsReference",
	$baseUrl: "http://localhost:8124/mobile1/x3/erp/SUPERV",
	$properties: {
		"SITE": {
			"$type": "application/x-string",
			"$maxLength": 5,
			"$isExcluded": true
		},
		"SITE_REF": {
			$type: "application/x-reference",
			$title: "Reference",
			$shortTitle: "Reference",
			$capabilities: "sort,filter",
			$item: {
				$url: "{$baseUrl}/FACILITY('{$key}')?representation=FACILITY.$lookup",
				$value: "{FCY}",
				"FCY": "{SITE}",
				"$key": "{FCY}",
				"$properties": {
					"FCY": {
						"$type": "application/x-string"
					}
				},
				"$links": {
					"$details": {
						"$title": "details",
						"$target": "",
						"$type": "application/json;vnd.sage=syracuse",
						"$url": "{$baseUrl}/FACILITY('{SITE}')?representation=FACILITY.$details",
						"$method": "GET"
					},
					"$lookup": {
						"$title": "lookup",
						"$type": "application/json;vnd.sage=syracuse",
						"$url": "{$baseUrl}/FACILITY?representation=FACILITY.$lookup",
						"$method": "GET"
					},
					"$query": {
						"$title": "query",
						"$target": "",
						"$type": "application/json;vnd.sage=syracuse",
						"$url": "{$baseUrl}/FACILITY?representation=FACILITY.$query",
						"$method": "GET"
					},
					"SIT1": {
						"$title": "SIT1",
						"$target": "main",
						"$type": "application/json;vnd.sage=syracuse",
						"$url": "https://maps.google.de/maps?q={SITE}&t=Default",
						"$method": "GET"
					}
				}
			}
		}
	},
	$relations: {},
	$articles: {
		$details: {
			"$layoutType": "stack",
			$items: [{
				$bind: "SITE_REF",
				$title: "Ref field 1"
			}, {
				$bind: "SITE_REF",
				$title: "Ref field 2"
			}, {
				$bind: "SITE_REF",
				$title: "Ref field 3"
			}, {
				$bind: "SITE_REF",
				$title: "Ref field 4"
			}, {
				$bind: "SITE_REF",
				$title: "Ref field 5"
			}, {
				$bind: "SITE_REF",
				$title: "Ref field 6"
			}, {
				$bind: "SITE_REF",
				$title: "Ref field 7"
			}]
		},
		$edit: {
			"$layoutType": "stack",
			$items: [{
				$bind: "SITE_REF",
				$title: "Ref field 1"
			}, {
				$bind: "SITE_REF",
				$title: "Ref field 2"
			}, {
				$bind: "SITE_REF",
				$title: "Ref field 3"
			}, {
				$bind: "SITE_REF",
				$title: "Ref field 4"
			}, {
				$bind: "SITE_REF",
				$title: "Ref field 5"
			}, {
				$bind: "SITE_REF",
				$title: "Ref field 6"
			}, {
				$bind: "SITE_REF",
				$title: "Ref field 7"
			}]
		}
	},
	$services: {
		$new: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve({});
			return deferred.promise();
		},
		$read: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve(_data);
			return deferred.promise();
		},
		$query: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve({
				$resources: [_data]
			});
			return deferred.promise();
		},
		$save: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve();
			return deferred.promise();
		},
		$delete: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve();
			return deferred.promise();
		}
	}
};
