// syracuse-tablet/html/js/sdata/entities/test/testGaugeChartDashboard.js
"use strict";

exports.dashboards = {
	"testGaugeChartDashboard": {
		"$dashboardName": "testGaugeChartDashboard",
		"$title": "Gauge chart test dashboard",
		"$description": "Test different gauge charts authoring",
		"$vignettes": {
			"testGauge": {
				"$uuid": "testGauge",
				"$displayStyle": "$full"
			}
		},
		"$article": {
			"$layoutType": "stack",
			"$items": [{
				"$layoutType": "row",
				"$widthXs": "3,3,3,3",
				"$items": [{
					"$bind": "testGauge",
					$article: {
						$layoutType: "stack",
						$items: [{
							$bind: "salesOp",
							/* GAUGE 1 */
							$gauge: {}
						}]
					}
				}, {
					"$bind": "testGauge",
					$article: {
						$layoutType: "stack",
						$items: [{
							$bind: "salesOp",
							/* GAUGE 2 */
							"$gauge": {
								"segments": [{
									"propStart": 0,
									"propEnd": 0.75
								}, {
									"propStart": 0.75,
									"propEnd": 0.85
								}, {
									"propStart": 0.85,
									"propEnd": 1
								}]
							}
						}]
					}
				}, {
					"$bind": "testGauge",
					$article: {
						$layoutType: "stack",
						$items: [{
							$bind: "salesValue",
							/* GAUGE 3 */
							"$gauge": {
								"$bindMin": "salesValueMin",
								"$bindMax": "salesValueMax",
								"segments": [{
									"propStart": 0,
									"propEnd": 0.75
								}, {
									"propStart": 0.75,
									"propEnd": 0.85
								}, {
									"propStart": 0.85,
									"propEnd": 1
								}]
							}
						}]
					}
				}, {
					"$bind": "testGauge",
					$article: {
						$layoutType: "stack",
						$items: [{
							$bind: "salesValue",
							/* GAUGE 4 */
							"$gauge": {
								"$bindMin": "salesValueMin",
								"$bindMax": "salesValueMax",
								"segments": [{
									"valStart": 100000,
									"valEnd": 130000,
									"color": "#DF5353"
								}, {
									"valStart": 130000,
									"valEnd": 140000
								}, {
									"valStart": 140000,
									"valEnd": 150000,
									"color": "#55BF3B"
								}]
							}
						}]
					}
				}]
			}]
		}
	}
};

exports.gadgets = {
	"testGauge": {
		"$type": "$representation",
		"$title": "Gauge 1",
		"entity": "testGaugeChart",
		"action": "$details",
		"representation": "testGaugeChart",
		"facet": "$details"
	}
};
