// syracuse-tablet/html/js/sdata/entities/test/testFormatsReference.js
"use strict";


var $descriptionSamples = ["Labe", "Labuan", "La Rochelle", "Las Vegas", "Las Palmas", "Liege", "Libreville", "Lille", "Liverpool", "Leipzig", "Lucknow", "London"];
var $codeSamples = ["LEK", "LBU", "LRH", "LAS", "LPA", "LGG", "LBV", "LIL", "LPL", "LEJ", "LKO", "LON"];

var _data = {
	SITE_REF: {
		$description: "Paris",
		FCY: "PAR"
	},
	SITE_REF_1: {
		$description: "Yaounde",
		FCY: "YDE"
	},
	SITE_REF_2: {
		$description: "London",
		FCY: "LON"
	},
	SITE_REF_3: {
		$description: "Lisbon",
		FCY: "LIS"
	}
};

var _qData = [];
var j = -1;
var m = -1;
for (var i = 0; i < 3; i++) {
	_qData.push({
		$uuid: i,
		SITE_REF: {
			$description: $descriptionSamples[++j],
			FCY: $codeSamples[++m]
		},
		SITE_REF_1: {
			$description: $descriptionSamples[++j],
			FCY: $codeSamples[++m]
		},
		SITE_REF_2: {
			$description: $descriptionSamples[++j],
			FCY: $codeSamples[++m]
		},
		SITE_REF_3: {
			$description: $descriptionSamples[++j],
			FCY: $codeSamples[++m]
		}
	});
}

exports.entity = {
	$entityName: "testFormatsReference",
	$baseUrl: "http://localhost:8124/mobile1/x3/erp/SUPERV",
	$properties: {
		"SITE": {
			"$type": "application/x-string",
			"$maxLength": 5,
			"$isExcluded": true
		},
		"SITE_REF": {
			$type: "application/x-reference",
			$title: "Reference",
			$shortTitle: "Reference",
			$capabilities: "sort,filter",
			$item: {
				$url: "{$baseUrl}/testFormatsReference('{$key}')?representation=testFormatsReference.$lookup",
				$value: "{FCY}",
				"FCY": "{SITE}",
				"$key": "{FCY}",
				"$properties": {
					"FCY": {
						"$type": "application/x-string"
					}
				},
				"$links": {
					"$details": {
						"$title": "details",
						"$target": "",
						"$type": "application/json;vnd.sage=syracuse",
						"$url": "{$baseUrl}/testFormatsReference('{SITE}')?representation=testFormatsReference.$details",
						"$method": "GET"
					},
					"$lookup": {
						"$title": "lookup",
						"$type": "application/json;vnd.sage=syracuse",
						"$url": "{$baseUrl}/testFormatsReference?representation=testFormatsReference.$lookup",
						"$method": "GET"
					},
					"$query": {
						"$title": "query",
						"$target": "",
						"$type": "application/json;vnd.sage=syracuse",
						"$url": "{$baseUrl}/testFormatsReference?representation=testFormatsReference.$query",
						"$method": "GET"
					},
					"SIT1": {
						"$title": "SIT1",
						"$target": "main",
						"$type": "application/json;vnd.sage=syracuse",
						"$url": "https://maps.google.de/maps?q={SITE}&t=Default",
						"$method": "GET"
					}
				}
			}
		},
		"SITE_REF_1": {
			$type: "application/x-reference",
			$title: "Reference 1",
			$shortTitle: "Reference 1",
			$capabilities: "sort,filter",
			$item: {
				$url: "{$baseUrl}/testFormatsReference('{$key}')?representation=testFormatsReference.$lookup",
				$value: "{FCY}",
				"FCY": "{SITE}",
				"$key": "{FCY}",
				"$properties": {
					"FCY": {
						"$type": "application/x-string"
					}
				},
				"$links": {
					"$details": {
						"$title": "details",
						"$target": "",
						"$type": "application/json;vnd.sage=syracuse",
						"$url": "{$baseUrl}/testFormatsReference('{SITE}')?representation=testFormatsReference.$details",
						"$method": "GET"
					},
					"$lookup": {
						"$title": "lookup",
						"$type": "application/json;vnd.sage=syracuse",
						"$url": "{$baseUrl}/testFormatsReference?representation=testFormatsReference.$lookup",
						"$method": "GET"
					},
					"$query": {
						"$title": "query",
						"$target": "",
						"$type": "application/json;vnd.sage=syracuse",
						"$url": "{$baseUrl}/testFormatsReference?representation=testFormatsReference.$query",
						"$method": "GET"
					},
					"SIT1": {
						"$title": "SIT1",
						"$target": "main",
						"$type": "application/json;vnd.sage=syracuse",
						"$url": "https://maps.google.de/maps?q={SITE}&t=Default",
						"$method": "GET"
					}
				}
			}
		},
		"SITE_REF_2": {
			$type: "application/x-reference",
			$title: "Reference 2",
			$shortTitle: "Reference 2",
			$capabilities: "sort,filter",
			$item: {
				$url: "{$baseUrl}/testFormatsReference('{$key}')?representation=testFormatsReference.$lookup",
				$value: "{FCY}",
				"FCY": "{SITE}",
				"$key": "{FCY}",
				"$properties": {
					"FCY": {
						"$type": "application/x-string"
					}
				},
				"$links": {
					"$details": {
						"$title": "details",
						"$target": "",
						"$type": "application/json;vnd.sage=syracuse",
						"$url": "{$baseUrl}/testFormatsReference('{SITE}')?representation=testFormatsReference.$details",
						"$method": "GET"
					},
					"$lookup": {
						"$title": "lookup",
						"$type": "application/json;vnd.sage=syracuse",
						"$url": "{$baseUrl}/testFormatsReference?representation=testFormatsReference.$lookup",
						"$method": "GET"
					},
					"$query": {
						"$title": "query",
						"$target": "",
						"$type": "application/json;vnd.sage=syracuse",
						"$url": "{$baseUrl}/testFormatsReference?representation=testFormatsReference.$query",
						"$method": "GET"
					},
					"SIT1": {
						"$title": "SIT1",
						"$target": "main",
						"$type": "application/json;vnd.sage=syracuse",
						"$url": "https://maps.google.de/maps?q={SITE}&t=Default",
						"$method": "GET"
					}
				}
			}
		},
		"SITE_REF_3": {
			$type: "application/x-reference",
			$title: "Reference 3",
			$shortTitle: "Reference 3",
			$capabilities: "sort,filter",
			$item: {
				$url: "{$baseUrl}/testFormatsReference('{$key}')?representation=testFormatsReference.$lookup",
				$value: "{FCY}",
				"FCY": "{SITE}",
				"$key": "{FCY}",
				"$properties": {
					"FCY": {
						"$type": "application/x-string"
					}
				},
				"$links": {
					"$details": {
						"$title": "details",
						"$target": "",
						"$type": "application/json;vnd.sage=syracuse",
						"$url": "{$baseUrl}/testFormatsReference('{SITE}')?representation=testFormatsReference.$details",
						"$method": "GET"
					},
					"$lookup": {
						"$title": "lookup",
						"$type": "application/json;vnd.sage=syracuse",
						"$url": "{$baseUrl}/testFormatsReference?representation=testFormatsReference.$lookup",
						"$method": "GET"
					},
					"$query": {
						"$title": "query",
						"$target": "",
						"$type": "application/json;vnd.sage=syracuse",
						"$url": "{$baseUrl}/testFormatsReference?representation=testFormatsReference.$query",
						"$method": "GET"
					},
					"SIT1": {
						"$title": "SIT1",
						"$target": "main",
						"$type": "application/json;vnd.sage=syracuse",
						"$url": "https://maps.google.de/maps?q={SITE}&t=Default",
						"$method": "GET"
					}
				}
			}
		}
	},
	$relations: {},
	$articles: {
		$details: {
			"$layoutType": "stack",
			$items: [{
				$bind: "SITE_REF",
				$title: "Ref field 1"
			}, {
				$bind: "SITE_REF_1",
				$title: "Ref field 2"
			}, {
				$bind: "SITE_REF_2",
				$title: "Ref field 3"
			}, {
				$bind: "SITE_REF_3",
				$title: "Ref field 4"
			}]
		},
		$edit: {
			"$layoutType": "stack",
			$items: [{
				$bind: "SITE_REF",
				$title: "Ref field 1"
			}, {
				$bind: "SITE_REF_1",
				$title: "Ref field 2"
			}, {
				$bind: "SITE_REF_2",
				$title: "Ref field 3"
			}, {
				$bind: "SITE_REF_3",
				$title: "Ref field 4"
			}]
		}
	},
	$services: {
		$new: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve({});
			return deferred.promise();
		},
		$read: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve(_data);
			return deferred.promise();
		},
		$query: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve({
				$resources: _qData
			});
			return deferred.promise();
		},
		$save: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve();
			return deferred.promise();
		},
		$delete: function(op, data) {
			var deferred = $.Deferred();
			deferred.resolve();
			return deferred.promise();
		}
	}
};
