// @sage/syracuse-react/src/over/modal/page/pageModal.tsx
import * as React from 'react';
import * as ReactDom from 'react-dom';
import * as Dom from '../../../utility/html/dom';
import * as Events from '../../../event/events';
import { Link } from '../../../link/link';
import Overlay from '../../overlay/overlay';
import { getLocalString } from '../../../culture/localString';
import { KeySpot } from '../../../shortcut/keySpot';

import '../modal.less';

const BODY_MARGIN = 48;

export interface IPageModalProps {
    deactivated: boolean, //classicPage compatibility
    onDidMount?: (modal: PageModal) => void,
    onResize?: (modal: PageModal) => void,
    onOk?: () => void,
    onClose: () => void,
    pageMode?: string,
    className?: string,
    title?: string
}

export class PageModal extends React.Component<IPageModalProps, {}> {
    keySpot: KeySpot;
    body: HTMLElement;
    dialog: HTMLElement;
    yMargin: number;
    xMargin: number;
    maxHeight: number;
    private _header: HTMLElement;
    private _footer: HTMLElement;

    get header(): HTMLElement {
        if (!this._header) {
            this._header = this.body.previousSibling as HTMLElement;
        }
        return this._header;
    }
    get footer(): HTMLElement {
        if (!this._footer) {
            this._footer = this.body.nextSibling as HTMLElement;
        }
        return this._footer;
    }
    get width(): number {
        return this.dialog.getBoundingClientRect().width;
    }
    set width(value: number) {
        this.dialog.style.width = value + "px";
    }
    componentDidMount() {
        this.props.onDidMount && this.props.onDidMount(this);
        Events.resize.subscribe(this.onResize);
        this.keySpot = new KeySpot("PageModal");
        this.keySpot.bindEscape(() => {
            if (!this.props.deactivated) {
                this.props.onClose();
                return true;
            }
        });
    }
    componentDidUpdate(prevProps: IPageModalProps) {
        if (this.props.deactivated != (prevProps && prevProps.deactivated)) {
            Dom.hide(ReactDom.findDOMNode(this) as HTMLElement, this.props.deactivated);
        }
    }
    onResize = () => {
        this.maxHeight = document.body.clientHeight * 0.9;
        this.yMargin = BODY_MARGIN;
        this.xMargin = BODY_MARGIN;
        this.maxHeight -= (this.yMargin + (this.header.getBoundingClientRect().height + this.footer.getBoundingClientRect().height));
        this.body.style.maxHeight = this.maxHeight + "px";
        this.props.onResize && this.props.onResize(this);
    }
    componentWillUnmount() {
        this.keySpot && this.keySpot.dispose();
        Events.resize.unsubscribe(this.onResize);
        this.body = this.dialog = this._header = this._footer = null;
    }
    render(): any {
        return <Overlay >
            <div className={(this.props.className ? this.props.className : "") + " s_modal_dialog"}
                data-s-dialog={this.props.pageMode}
                data-s-dialog-modal="true"
                ref={(node) => this.dialog = node}>
                <article className="s_modal_content">
                    <div className="s_modal_header">
                        {this.props.title}
                        <Link
                            className="s_modal_close"
                            icon={"close"}
                            title={getLocalString("diag_close")}
                            onClick={this.props.onClose} />

                    </div>
                    <div className="s_modal_body"
                        ref={(node) => this.body = node}>
                        {this.props.children}
                    </div>
                    <footer className={"s_modal_footer"}>
                        <Link
                            className="s_modal_footer_link"
                            title={this.props.onOk ? getLocalString("over_cancel") : getLocalString("over_close")}
                            onClick={this.props.onClose} />
                        {this.props.onOk && <Link
                            className="s_modal_footer_link OkLink"
                            title={getLocalString("over_ok")}
                            onClick={this.props.onOk} />}
                    </footer>
                </article>
            </div>
        </Overlay>;
    }
}
