// @sage/syracuse-lib/src/pdf/utils.ts
"use strict";

export function removeUnnecessaryBlank(str) {
	return str.replace(/\s*(\<\<|\>\>|\/|\[|\]|\(|\))\s*/g, function(s, sub) {
		return sub;
	});
};

export function formatDate(date) {
	//return a string like (D:20120320170324)
	var year = date.getUTCFullYear(),
		month = ("00" + (date.getUTCMonth() + 1)).slice(-2),
		day = ("00" + date.getUTCDate()).slice(-2),
		hours = ("00" + date.getUTCHours()).slice(-2),
		minutes = ("00" + date.getUTCMinutes()).slice(-2),
		seconds = ("00" + date.getUTCSeconds()).slice(-2);
	return "(D:" + year + month + day + hours + minutes + seconds + "Z)";
};

export function startsWith(str, pattern) {
	return str.substring(0, pattern.length) === pattern;
};

export function endsWith(str, pattern) {
	return str.substring(str.length - pattern.length) === pattern;
};

export function extractCertificateContent(str) {
	var begin = '-----BEGIN CERTIFICATE-----';
	var end = '-----END CERTIFICATE-----';
	return str.substring(str.indexOf(begin) + begin.length, str.indexOf(end)).replace(/\s/g, '');
};

export function error(msg) {
	return new Error(msg);
};
