// syracuse-tablet/html/js/sdata/sdataSyncUtils.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var sdataUtils = require('syracuse-tablet/html/js/sdata/sdataUtils');
var globals = require('syracuse-tablet/html/js/helpers/globals');
var prototypeModule = require('syracuse-tablet/html/js/helpers/prototype');

var _syncMaxLevels = 5;

exports.getSyncStructure = function(prototype) {
	var node = {
		$title: prototype.data("$title")
	};
	return _resolvePrototype({
		prototype: prototype,
		level: 1,
		maxLevels: _syncMaxLevels
	}, node);
};

function _resolvePrototype(opts, node) {

	var $links = opts.prototype.data("$links");

	if (opts.prototype.getFacet() === "query") {
		node.$syncType = "query";
		var $resources = opts.prototype.property("$resources");
		if ($resources && $resources.$item && $resources.$item.$links) {
			$links = $.extend({}, $links, $resources.$item.$links);
		}
	} else {
		node.$syncType = "details";
	}

	if (opts.level >= opts.maxLevels) {
		return $.smResolve(node);
	}
	node.$children = [];
	$.smForEachPromise(Object.keys($links), function(name) {
		var link = $links[name];
		if (_isCacheableLink(link, name)) {
			//var url = sdataUtils.parseSDataURL(link.$url);
			var repr = _getRepresentation(link);
			if (repr) {
				var child = {
					$name: name,
					$title: utils.parseExpression(link.$title, null, opts.prototype, true),
					$link: link,
					$representation: repr
				};
				node.$children.push(child);
				return _processNode(child, opts);
			}
		}
		return $.smResolve();
	});
	return $.smResolve(node);
}

function _isCacheableLink(link, name) {
	if (name.indexOf("$") === 0 && name !== "$details") {
		return false;
	}
	if (link.$method !== "GET") {
		return false;
	}
	if (link.$url.indexOf("/$services/") > -1) {
		return false;
	}
	if (link.$url.indexOf("representation=") < 0) {
		return false;
	}
	return true;
}

function _getRepresentation(link) {
	var repr = /[\?\&]representation=(\S+?)(&|$)/.exec(link.$url);
	return repr && repr[1];
}

function _processNode(node, opts) {
	var meta = globals.getMetaData();
	var protoName = _getProtoName(node, opts);
	return meta.getPrototype(protoName)
		.then(function(proto) {
			if (proto) {
				proto = prototypeModule.create(proto);

				if (node.$name === "$details") {
					node.$title = proto.data("$title");
				}

				return _resolvePrototype({
					prototype: proto,
					level: opts.level + 1,
					maxLevels: opts.maxLevels
				}, node);
			}
		});
}

function _getProtoName(node, opts) {
	var endpoint = opts.prototype.data("$baseUrl");
	endpoint = endpoint.split("/");
	endpoint = endpoint.slice(endpoint.length - 3, endpoint.length).join(".");
	return endpoint + "." + node.$representation;
}
