// syracuse-tablet/html/js/sdata/sdataCommonResources.js
"use strict";

var sdataUtils = require('syracuse-tablet/html/js/sdata/sdataUtils');
var sdataHttp = require('syracuse-tablet/html/js/sdata/sdataHttp');
var utils = require('syracuse-tablet/html/js/helpers/utils');
var locale = require('syracuse-tablet/html/js/helpers/locale');

/*
 * This file should be the only file containing direct, non generic access to server side supporting entities/services
 */
var _appListUrl = "/sdata/syracuse/collaboration/syracuse/mobileApplications/$service/availableApplications";
var _localePreferencesUrl = "/sdata/syracuse/collaboration/syracuse/localePreferences?representation=localePreference.$bulk";
var _userProfileUrl = "/sdata/syracuse/collaboration/syracuse/userProfiles/$service/current";
var _endpointListUrl = "/sdata/syracuse/collaboration/syracuse/endPoints?representation=endPoint.$bulk";
var _roleListUrl = "/sdata/syracuse/collaboration/syracuse/roles?representation=role.$bulk";

function _generateReject(title, result, url) {
	if (result && result.responseJSON && result.responseJSON.$diagnoses) {
		// $noConnectionError to notify caller that fail is due to network connection broke down
		return $.smReject({
			$noConnectionError: result.responseJSON.$noConnectionError,
			$title: title,
			$diagnoses: result.responseJSON.$diagnoses
		});
	}
	return $.smReject({
		$title: title
	});
}

exports.queryAllApplications = function() {
	var params = {
		method: "GET",
		url: _appListUrl
	};
	params.parsedUrl = sdataUtils.parseSDataURL(params.url);
	return sdataHttp.send(params)
		.then(function(result) {
			if (result.isSuccess && result.responseJSON) {
				var apps = result.responseJSON.$resources;
				return apps;
			}
			return _generateReject(locale.text("err.load.applst"), result, params.url);
		});
};

exports.queryApplicationDetail = function(appHeader, endpoint) {
	var params = {
		method: "GET",
		url: appHeader.$metaDataUrl + "?endpoint=" + endpoint
	};
	params.parsedUrl = sdataUtils.parseSDataURL(params.url);
	return sdataHttp.send(params)
		.then(function(result) {
			if (result.isSuccess && result.responseJSON) {
				var app = result.responseJSON;
				return app;
			}
			return _generateReject(locale.text("err.load.appdetail", [appHeader.applicationName]), result, params.url);
		});
};

exports.getUserProfile = function() {
	var params = {
		method: "GET",
		url: _userProfileUrl
	};
	params.parsedUrl = sdataUtils.parseSDataURL(params.url);
	return sdataHttp.send(params)
		.then(function(result) {
			if (result.isSuccess && result.responseJSON) {
				var profile = result.responseJSON;
				return profile;
			}
			return _generateReject(locale.text("err.load.profile"), result, params.url);
		});
};
/**
 * noFailIfNoConnection	see _fetchAll
 */
exports.queryLocalePreferences = function(noFailIfNoConnection) {
	return _fetchAll(_localePreferencesUrl, noFailIfNoConnection)
		.then(function(result) {
			var locales = [];
			result.forEach(function(locale) {
				locales.push({
					"$uuid": locale.$uuid,
					"code": locale.code,
					"description": locale.description
				});
			});
			return locales;
		});
};
/**
 * noFailIfNoConnection	see _fetchAll
 */
exports.queryEndpoints = function(noFailIfNoConnection) {
	return _fetchAll(_endpointListUrl, noFailIfNoConnection);
};
/**
 * noFailIfNoConnection	see _fetchAll
 */
exports.queryRoles = function(noFailIfNoConnection) {
	return _fetchAll(_roleListUrl, noFailIfNoConnection);
};

exports.updateUserProfile = function(changes) {

	function createWC() {
		return $.smResolve().then(function() {
			var trackingId = utils.UUID();
			var params = {
				method: "POST",
				url: "/sdata/syracuse/collaboration/syracuse/userProfiles/$template/$workingCopies?representation=userProfile.$edit&role=&trackingId=" + trackingId
			};
			params.parsedUrl = sdataUtils.parseSDataURL(params.url);
			return sdataHttp.send(params)
				.then(function(result) {
					if (result.isSuccess && result.responseJSON) {
						return $.smResolve(result.responseJSON);
					} else {
						return $.smReject();
					}
				});
		});
	}

	function updateWC(up) {
		return $.smResolve().then(function() {
			var upData = $.extend({}, up, changes);
			var params = {
				method: "PUT",
				url: up.$url,
				send: upData
			};
			params.parsedUrl = sdataUtils.parseSDataURL(params.url);
			return sdataHttp.send(params)
				.then(function(result) {
					if (result.isSuccess && result.responseJSON) {
						return $.smResolve(result.responseJSON);
					} else {
						return $.smReject();
					}
				});
		});
	}

	function deleteWC(up) {
		return $.smResolve().then(function() {
			var params = {
				method: "DELETE",
				url: up.$url
			};
			params.parsedUrl = sdataUtils.parseSDataURL(params.url);
			return sdataHttp.send(params)
				.then(function(result) {
					return $.smResolve();
				});
		});
	}

	return createWC()
		.then(function(up) {
			return updateWC(up);
		})
		.then(function(up) {
			return deleteWC(up).always(function() {
				return $.smResolve();
			});
		});
};
/**
 * noFailIfNoConnection	true to resolve with an empty list if httpstatus 0 (no network connection)
 * if noFailIfNoConnection == true and a network connection occurred -> Resolve with [] and true
 */
function _fetchAll(urlQuery, noFailIfNoConnection) {
	var deferred = $.Deferred();
	var list = [];

	function _fetchPage(url) {
		var params = {
			method: "GET",
			url: url
		};

		params.parsedUrl = sdataUtils.parseSDataURL(params.url);
		sdataHttp.send(params)
			.then(function(result) {
				var data = result && result.responseJSON;
				if (data && data.$noConnectionError === true && noFailIfNoConnection === true) {
					deferred.resolve([], true);
					return;
				}
				if (result && result.status !== 200) {
					var message;
					if (result.status === 401) {
						message = locale.text("error.session.timeout");
					} else if (data && data.$diagnoses && data.$diagnoses.length > 0) {
						message = data.$diagnoses[0].$message;
					} else {
						message = locale.text("fetchurl.error", [params.parsedUrl.contract + "/" + params.parsedUrl.entity, result.status]);
					}
					deferred.reject({
						$message: message,
						$redirectPage: "login"
					});
				}
				data && data.$resources && data.$resources.forEach(function(i) {
					list.push(i);
				});
				if (data && data.$links && data.$links.$next && data.$links.$next.$url) {
					// Scan all resources
					_fetchPage(data.$links.$next.$url);
				} else {
					// Resolve in no $links 
					deferred.resolve(list, false);
				}
			})
			.fail(function(e) {
				deferred.reject(e);
			});
	}

	_fetchPage(urlQuery);
	return deferred.promise();
}
