// syracuse-tablet/html/js/sdata/sdataUtils.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var globals = require('syracuse-tablet/html/js/helpers/globals');


var _rID = /^(.*)\(\'(.*)\'\)$/i;
var _parseSDataURL = function(url) {
	var res = utils.parseURL(url);
	if (res.query && res.query.representation) {
		res.representation = res.query.representation;
		res.facet = null;
		var p = res.representation.indexOf('.');
		if (p != -1) {
			res.facet = res.representation.substr(p + 1);
			if (res.facet.charAt(0) == '$') res.facet = res.facet.substr(1);
			res.representation = res.representation.substr(0, p);
		}
	}
	var a = res.directory.split('/');
	if (a.length <= 4) {
		throw new Error("Bad sdata url:" + url);
	}
	res.dispatch = a[1];
	res.application = a[2];
	res.contract = a[3];
	res.dataset = a[4];
	if (a.length === 4) return res;
	var match = _rID.exec(a[5]);
	if (match && (match.length == 3)) {
		res.entity = match[1];
		res.id = match[2];
		if (res.id && res.id.trim().Length === 0) res.id = null;
	} else {
		res.entity = a[5];
		res.id = null;
	}
	if (a.length === 5) return res;
	return res;
};

/**
 * Resolve sdataurl and return navigation info to generate the <a>/<button> link
 */
var _getLinkInfo = function(sDataUrl, dao, acceptNoRep) {
	var uu = sDataUrl;
	if (dao) { // dao is optional since a link can be a link without dataset (e.g. prev/next)
		sDataUrl = dao.parseExpression(sDataUrl);
	}
	var url = _parseSDataURL(sDataUrl);
	if (url.query.representation || acceptNoRep === true) {
		var ep = url.application + "." + url.contract + "." + url.dataset;
		return {
			page: ep + (url.query.representation ? "." + url.query.representation : ""),
			sDataUrl: sDataUrl
		};
	} else if (url.entity === "$mobileDashboards") {
		var app = globals.getApplication();
		var dashboards = app.dao && app.dao.$dashboards || [];
		var match = false;
		if (!Object.keys(dashboards).some(function(dashboard) {
			return (dashboards[dashboard].$dashboardName === url.query.dashboard);
		})) {
			throw new Error("Cannot find dashboard: " + url.query.dashboard);
		}

		return {
			page: url.query.dashboard + ".$mobileDashboard",
			parameters: url.query.parameters
		};
	} else {
		throw new Error("getLinkInfo - Unexpected url" + sDataUrl);
	}
};
/**
 * Returns attributes for link tag
 * 	linkInfo: 	object returned by _getLinkInfo
 * 				or sdataUrl
 * dao:			mandatory if linkInfo is a sdataUrl
 */
var _getLinkAttrs = function(linkInfo, dao) {
	if (typeof linkInfo === "string") {
		// We assume that linkInfo is the url - we need dao
		linkInfo = _getLinkInfo(linkInfo, dao);
	}
	if (linkInfo.type === "navig") {
		return 'data-nav="' + linkInfo.page + '" data-sdata-url="' + encodeURIComponent(linkInfo.sDataUrl) + '"';
	} else if (linkInfo.type === "act") {
		return 'data-action="' + linkInfo.action + '" data-params="' + encodeURIComponent((linkInfo.params || "")) + '"';
	} else {
		throw new Error("_getLinkAttrs - unknonwn type\n" + linkInfo.type);
	}
};
/**
 * SCan diagnoses in a sdata response
 * TODO TEST
 */
var _clearSyracuseStack = function(d) {
	if (d.$stackTrace && d.$stackTrace.indexOf("node_modules\syracuse-x3") >= 0) {
		if (d.$message && d.$stackTrace.indexOf(d.$message) >= 0) {
			// No need of stack
			delete d.$stackTrace;
		} else {
			// Remove stack because too long
			d.$stackTrace = d.$stackTrace.split("\n")[0];
		}
	}
	return d;
};
var _scanDiagnoses = function(data) {
	var result = [];
	var done = [];
	var _scan = function(o, prop, path) {
		if (!path) path = [];
		if (done.indexOf(o) >= 0) return;
		if (Array.isArray(o)) {
			done.push(o);
			if (prop) path.push(prop);
			o.forEach(function(e, idx) {
				_scan(e, idx + "", path);
			});
			if (prop) path.pop(prop);
		} else if ($.isPlainObject(o)) {
			done.push(o);
			for (var p in o) {
				if (p === "$diagnoses") {
					o[p].forEach(function(d) {
						var $path = [].concat(path);
						if (prop) $path.push(prop);
						if (p) $path.push(p);
						d = _clearSyracuseStack($.extend({}, d));
						d.$path = $path.join('.');
						result.push(d);
					});
				} else {
					if (prop) path.push(prop);
					_scan(o[p], p, path);
					if (prop) path.pop(prop);
				}
			}
		}
	};
	if (data != null) _scan(data);
	return result;

};

var _addWhereClause = function(url, where, params) {
	params = params || {};
	var allParamsSet = true;
	where = where.value;
	var res = where.replace(/\{(\w*?)\}/g, function(m, p1) {
		var value = params[p1];
		if (value == null) {
			allParamsSet = false;
		}
		return value != null ? value : "";
	});
	if (allParamsSet) {
		url = url + "&where=" + res;
	}
	return url;
};

var _displayActionMsg = function(success, name, data, cb) {
	var obj;
	var diags = _scanDiagnoses(data);
	if (diags.length > 0) {
		obj = {
			$diagnoses: diags
		};
	} else {
		obj = {
			$diagnoses: [{
				$message: "Action: " + name + " " + (success ? "succeeded" : "failed"),
				$severity: success ? "success" : "error",
				$stackTrace: JSON.stringify(data)
			}]
		};
	}
	globals.getModal().actionResult("Action: " + name, obj);
};

exports.parseSDataURL = _parseSDataURL;
exports.getLinkInfo = _getLinkInfo;
exports.getLinkAttrs = _getLinkAttrs;
exports.scanDiagnoses = _scanDiagnoses;
exports.addWhereClause = _addWhereClause;
exports.displayActionMsg = _displayActionMsg;
