// syracuse-tablet/html/js/sdata/sdataDispatcher.js
"use strict";

var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("sdata");
var sdataHttp = require('syracuse-tablet/html/js/sdata/sdataHttp');
var sdataUtils = require('syracuse-tablet/html/js/sdata/sdataUtils');

/**
 * link
 * {
 * 	 $url: ...,
 * 	 $method: "GET", ...
 * }
 * data
 * {
 *   ...
 * }
 */
exports.dispatch = function(link, data) {
	var deferred = $.Deferred();
	var send = {
		method: link.$method || "GET",
		url: link.$url
	};

	if (link.$method === "PUT" || link.$method === "POST") {
		send.data = data;
	}

	//	if (params.parsedUrl.dispatch === "sdata") {
	//		
	//	}
	sdataHttp.send(send)
		.then(function(result) {

		});
	return deferred.promise();
};

/**
 * Parse link to identify which kind of operation it's intended to start
 *
 * link:
 * {
 * 	 $url: ...,
 * 	 $method: "GET", ...
 * }
 */
exports.parseLink = function(link) {
	var url = link.$url;
	var result = {
		$link: link,
		$parsedUrl: sdataUtils.parseSDataURL(link.$url)
	};
	var dir = result.$parsedUrl.directory; // All except query parameters
	var idx;
	if (link.$method === "GET") {
		if (url.indexOf("$prototypes") > -1) {
			result.$operation = "prototype";
			result.$representation = /\$prototypes\('(.*?)'\)/.exec(url)[1];
		} else if (dir.indexOf("$template") > -1) {
			result.$operation = "new";
			result.$representation = result.$parsedUrl.query && result.$parsedUrl.query.representation;
		} else if ((idx = dir.indexOf("/$services/")) > -1) {
			/*
			var service = dir.substring(idx + 11);
			result.$operation = "service";
			result.$service = service;
			result.$representation = result.$parsedUrl.query && result.$parsedUrl.query.representation;
			*/
		} else {
			var key = /\((.*?)\)/.exec(dir);
			if (key) {
				result.$operation = "read";
				result.$representation = result.$parsedUrl.query && result.$parsedUrl.query.representation;
			} else {
				result.$operation = "query";
				result.$representation = result.$parsedUrl.query && result.$parsedUrl.query.representation;
			}
		}
	} else if ((idx = dir.indexOf("/$services/")) > -1) {
		var service = dir.substring(idx + 11);
		result.$operation = "service";
		result.$service = service;
		result.$representation = result.$parsedUrl.query && result.$parsedUrl.query.representation;
	} else if (link.$method === "DELETE") {
		var key = /\('(.*?)'\)/.exec(dir);
		result.$operation = "delete";
		result.$key = key[1];
		result.$representation = result.$parsedUrl.query && result.$parsedUrl.query.representation;
	} else {
		// PUT/POST and no service
		var key = /\('(.*?)'\)/.exec(dir);
		if (key) {
			result.$operation = "save";
			result.$key = key[1];
			result.$representation = result.$parsedUrl.query && result.$parsedUrl.query.representation;
		} else {
			result.$operation = "save";
			result.$representation = result.$parsedUrl.query && result.$parsedUrl.query.representation;
		}
	}

	return result;
};
