// @sage/syracuse-ui/lib/page/desktopPage.js
"use strict";
var _helpers = require('@sage/syracuse-core').helpers;
var _garbage = require('@sage/syracuse-ui/lib/page/tools/fieldsGarbage');
var _paging = require("@sage/syracuse-ui/lib/field/array/tools/paging");
var _designerViewChoice = require('@sage/syracuse-ui/lib/authoring/designerViewChoice');
var SideBar = require("@sage/syracuse-ui/lib/page/tools/sideBar").SideBar;

var _menus = {
	_mainActions: {
		$cube: "$create",
		$query: "$create",
		$details: "$edit",
		$edit: "$save",
		$request: "$invoke"
	},
	addToPage: function(page) {
		if (!page.$item.$menus) {
			page.$item.$menus = [{
				$skin: page.$skinMenu || "s-mn-main",
				$category: "menus",
				$isBoxCollapsable: true,
				$isMenusBag: true,
				$layout: {
					$items: []
				}
			}];
			if (page.$mainAction) {
				page.$item.$menus[0].$layout.$items.push({
					$showIfSet: true,
					$bind: page.$mainAction,
					$style: "main"
				});
			}
		} else {
			if (!Array.isArray(page.$item.$menus)) {
				page.$item.$menus = [{
					$isMenusBag: true,
					$layout: {
						$items: syra_pageBuilder.getDefinedFieldList(page.$item.$menus.$layout.$items)
					}
				}];
			}
		}
		for (var ii = 0, jj = page.$item.$menus.length; ii < jj; ii++) {
			var $menus = page.$item.$menus[ii];
			$menus.$category = "menus";
			if ($menus.$isBoxCollapsable === undefined) {
				$menus.$isBoxCollapsable = true;
			}
			if ($menus.$skin === undefined) {
				$menus.$skin = page.$skinMenu || "s-mn-main";
			}
		}
		return page.$item.$menus;
	},
	addBar: function(page) {
		var bar = page.menuBar = new SideBar();
		bar.preferenceKey = page.isVignettePage ? "vignette-menubar" : "menuBar";
		bar.options = bar.options || {};
		bar.options.leftSide = !!(page.$pageCategory == "vignette" || page.isLandingPage);
		bar.options.$skin = page.$skin + "-menubar";
		syra_item.initialize(page, bar);
		bar.slot = syra_dom.div(page.$skin + "-menubar-slot");
		bar.body = syra_dom.div(page.$skin + "-menubar-body");

		if (bar.options.leftSide) {
			bar.page.domItem.insertBefore(bar.slot, bar.page.domItem.firstChild);
		} else {
			bar.page.domItem.appendChild(bar.slot);
		}
		page.menuBar.load(page);
		if (!syra_site.mobile) {
			if (page.isFusionPage && page.$item.$menus && page.$item.$menus.length) {
				page.$item.$menus[0].$isMenusBag = true;
			}
			page.$mainAction = this._mainActions[page.$facet];
			page.menuBarLayout = page.addItem(bar.body, {
				$layoutType: "stack",
				$items: _menus.addToPage(page)
			});
		}
	}
};


function ResizableList() {
	this.stack = [];
}

_helpers.defineClass(ResizableList, null, {
	add: function(item) {
		this.stack.indexOf(item) < 0 && this.stack.push(item);
	},
	remove: function(item) {
		var ii = this.stack && this.stack.indexOf(item);
		(ii >= 0) && this.stack.splice(ii, 1);
	},
	resize: function(articleParent) {
		for (var ii = 0, jj = this.stack.length; ii < jj; ii++) {
			var item = this.stack[ii];
			if (!item.disposed && item.resizeItem && (item.displayed !== false)) {
				if (articleParent && (item.articleParent != articleParent)) {
					continue;
				}
				item.resizeItem();
			}
		}
	}
});

function DesktopPage() {}

exports.DesktopPage = _helpers.defineClass(DesktopPage, null, {
	getTitle: function(applyDefault) {
		var title = (this.domTitle && this.domTitle.textContent) || (this.$item && this.$item.$title) || "";
		if ((title == "" || title == "-") && applyDefault !== false) {
			title = this._defaultTitle || "";
		}
		return title;
	},
	getMainAction: function() {
		var $binds = Object.keys(this.menuItems);
		for (var ii = 0, jj = $binds.length; ii < jj; ii++) {
			var menus = this.menuItems[$binds[ii]];
			if (menus && menus.length && ((menus[0].$item.$style === "main") || (menus[0].$item.$kind === "main"))) {
				return menus[0];
			}
		}
	},
	ensureMenuBar: function() {
		if (!this.isMenuBarDisabled) {
			this.isMenuBarDisabled = this.$item.$isMenuBarDisabled;
		}!this.isMenuBarDisabled && _menus.addBar(this);
	},
	onItemInOut: function(onEnter, event, target) {},
	setTitle: function(title, isDynamic) {
		if (this.domTitle) {
			if (isDynamic && this.$item.$XID) {
				//set only by convergence setDataContainer
				this.dynamicTitles = this.dynamicTitles || {};
				this.dynamicTitles[this.$item.$XID] = title;
			}
			if (title != null && title != "") {
				this.titleText = title;
				if (this.titleText == "-") {
					this.titleText = "";
				}
				if (this.titleText.indexOf("{") < 0) {
					this.domTitle.textContent = this.titleText;
					if (!this.domTitle.syrainout) {
						this.domTitle.title = this.titleText;
					}
				} else {
					this.domTitle.textContent = "";
					var parsedText = syra_expression.render(this, this.titleText, this.domTitle, "s_page_title_field", false);
					if (parsedText && parsedText.charAt(0) == "@") {
						this.isTitleUnlocalized = this.titleText == ("{" + parsedText + "}");
						this.domTitle.style.visibility = this.isTitleUnlocalized ? "hidden" : "";
					} else {
						this.titleText = parsedText;
						if (this.isTitleUnlocalized) {
							this.isTitleUnlocalized = false;
							this.domTitle.style.visibility = "";
						}
					}
				}
				this.isTitleHidden = this.$isTitleHidden || this.$item.$isTitleHidden;
			} else {
				if (!this.isFusionPage) {
					this.isTitleHidden = true;
				} else {
					this.domTitle.textContent = this.titleText = "";
				}
			}
			var $description = this.dataset.$description || this.$prototype.$description || this.$item.$description;
			if ($description && this.titleText != $description) {
				if (!this.domDescription) {
					this.domDescription = syra_dom.div("s_page_description");
				}
				if ($description.indexOf("{") < 0) {
					this.domDescription.textContent = $description;
				} else {
					syra_expression.render(this, $description, this.domDescription, "s_page_description_field", false);
				}
				if (this.titleText != this.domDescription.textContent) {
					this.domTitleBar.appendChild(this.domDescription);
				}
			}
			syra_dom.hide(this.header, this.isTitleHidden);
		}
	},
	switchDesigner: function(open) {
		var self = this;
		if (open) {
			require.async("@sage/syracuse-ui/lib/authoring/pageDesigner", function(error, module) {
				if (error) {
					syra_alert.error(error);
					return;
				}
				syra_fields.advancedState.toggleAllFields(self, true);
				self.designer = new module.Designer();
				self.designer.openDesigner(self);
				syra_button.disable(self.authoringBtn, true);
				self.resizeItem(true);
			});
		} else {
			if (self.designer) {
				syra_alert.clear(self.page);
				syra_button.disable(self.authoringBtn, false);
				self.designer.dispose();
				syra_fields.advancedState.toggleAllFields(self);
				self.designer = null;
				if (self.isVignettePage) {
					self.resizeItem(true);
				} else {
					syra_site.resizeItem();
				}

			}
		}
	},
	setShortcutSpyerList: function(list) {
		if (list && !list.isArrayField) {
			while (list && !list.isArrayField) {
				list = list.articleParent;
			}
			if (!list) {
				return;
			}
		}
		this.shortcutSpyerList = list;
	},
	renderLayoutContent: function() {
		this.layoutContent = this.addItem(this.body, this.$item.$layout);
	},
	addItem: function(slot, $item, boxParent, initData) {
		var item = syra_item.add(this, $item, boxParent || this);
		if (item) {
			item.layoutSlot = slot;
			item.load(initData);
		}
		return item;
	},
	getPageSize: function(updated) {
		if (this.layoutSlot && this.size == undefined) {
			var slot = this.layoutSlot;
			var sizeSize = syra_site.getSize();
			var diff = 0;
			var isHFit = !this.inlinePageHost && !this.vignetteField;
			if (this.vignetteField && this.vignetteField.isMaximized) {
				isHFit = true;
				slot = this.vignetteField.page.layoutSlot;
				diff = this.vignetteField.header.clientHeight + (2 * this.vignetteField.header.offsetTop); //2* offsetTop for margin
			}
			if (slot == syra_site.siteApp) {
				var body = sizeSize.body;
				this.size = {
					height: body.height,
					width: body.width
				};
			} else {
				this.size = {
					height: this.layoutSlot.clientHeight,
					width: this.layoutSlot.clientWidth
				};
			}
			this.size.isHFit = isHFit;
			this.size.height = this.size.height - diff;
			this.size.headerHeight = this.header ? this.header.clientHeight : 0;
		}
		return this.size;
	},
	resizeItem: function(resize, articleParent) {
		if (this.displayed !== false && this.isPageLoaded && !this._isDataChanging) {
			if (syra_dom.isVisible(this.domItem)) {
				this.isResizing = true;
				delete this.size;
				this.getPageSize();
				this.designer && this.designer.resizeItem(resize);
				if (this.bars) {
					for (var ii = 0, jj = this.bars.length; ii < jj; ii++) {
						this.bars[ii].resizeSplitter();
					}
				}
				this.resizableList.resize(articleParent);
				this.inlinePageHost && this.inlinePageHost.onInlinePageResized(this);
				this.isResizing = false;
			}
		}
	},
	startChange: function() {
		this._isDataChanging = true;
	},
	applyChange: function(newData) {
		if (!this.disposed && this.dataset) {
			if (newData && newData.$isPartialDelta) {
				this.$isPartialDelta = true;
				//delete newData.$isPartialDelta;
			}
			if (newData) {
				syra_dataset.applyDelta(this, this.dataset, newData);
				newData.$description && this.setTitle(this.titleText, false, newData.$description);
				syra_article.applyChange(this, newData);
			}
			if (!this._isDataChanging) {
				this.endChange();
			}
			delete this.$isPartialDelta;
		}
	},
	endChange: function(discardRedraw) {
		this._isDataChanging = false;
		if (!this.disposed) {
			if (!discardRedraw) {
				if (this.menuBar && !this.menuBar.isInitialized) {
					this.menuBar.isInitialized = true;
					var body = (this.menuBarLayout && this.menuBarLayout.domItem) || this.menuBar.body;
					!body.firstChild && this.menuBar.toggleBar(false);
				}
				this.alertPanel && this.alertPanel.refresh();
				this.isPageLoaded && this.resizeItem(true);
			}
		}
		syra_quality.page_endChange(this);
	},
	isInMenuBar: function(item) {
		var menuGroupRoot = item.menuGroupRoot || (item.boxParent && item.boxParent.menuGroupRoot);
		var layoutParent = menuGroupRoot && menuGroupRoot.layoutParent;
		return layoutParent && layoutParent == this.menuBarLayout;
	},
	savePageDesign: function() {
		if (this.isVignettePage) {
			var $vignette = _helpers.object.clone(this.$item, true);
			delete $vignette.$menus;
			if (this.altVignetteKey) {
				(this.vignetteField.$item.$altVignettes = this.vignetteField.$item.$altVignettes || {})[this.altVignetteKey] = $vignette;
			} else {
				this.vignetteField.$item.$vignette = $vignette;
			}
			if (this.vignetteField.page.designer) {
				this.vignetteField.page.designer.endArticleUpdate();
			}
		}
	},
	load: function(initData) {
		this.isDesktopPage = true;
		switch (this.$pageCategory) {
			case "lookup":
				this.$skin = this.$item.$skin || "s-page";
				this._defaultTitle = syra_local.dskpPageTitle;
				break;
			default:
				this.$skin = this.$skin || this.$item.$skin || "s-page";
				if (this.inlinePageHost) {
					this.$skin = "s-inline-page " + this.$skin;
				}
				this._defaultTitle = this._defaultTitle || syra_local.dskpPageTitle;
				break;
		}
		var autoInsertFields;
		if (!this.isAutoInsertFieldDisabled) {
			autoInsertFields = this.insertNewItems();
		}

		this.bars = [];
		this.resizableList = new ResizableList();
		syra_pageBuilder.initialize(this);
		this.$skin = this.$item.$skin || this.$skin || "s-page";
		this.$designLevel = "article";
		this.externalAdapter = this.$item.externalAdapter || syra_site.externalAdapter;
		delete this.$item.externalAdapter;

		syra_article.beforeDraw(this);
		this.drawPage();
		syra_article.endDraw(this, initData);

		if (!this.isAutoInsertFieldDisabled) {
			(this.garbage = new _garbage.FieldsGarbage()).load(this);
			autoInsertFields && autoInsertFields.alert();
		}
		return this;
	},
	ensureSkin: function() {
		this.$skin = this.$skin || this.$item.$skin;
	},
	drawPage: function() {
		this.$item.$title = this.$item.$title || "{$title}";
		//this.domItem = syra_dom.article(this.$skin, this.layoutSlot);
		this.domItem = syra_dom.article(this.$skin);
		this.domItem.syraPage = this.domItem.syraItem = this.id;
		this.contentSlot = syra_dom.div(this.$skin + "-content-slot");
		this.alertPanel = syra_alert.addPanel(syra_dom.div("s_alertpanel_slot", this.contentSlot));
		this.scrollview = syra_dom.div("s_page_scrollview", this.contentSlot);

		this.body = syra_dom.div("s_page_body");

		this.header = syra_dom.header("s_page_head", this.scrollview);

		if (!this.isAlertModal && this.$breadCrumb && this.$breadCrumb.length > 0) {
			this.breadcrumb = syra_site.addBreadcrumPath(this);
		}
		this.headerBottom = syra_dom.div("s_page_head_bottom", this.header);

		this.domTitleBar = syra_dom.div("s_page_title_bar", this.headerBottom);
		this.domTitle = syra_dom.div("s_page_title", this.domTitleBar);
		this.domTitle.syraItem = this.id;
		delete this.$item.$isTitleHidden;
		this.setTitle(this.$item.$title || (this.$field ? this.$field.$title : null));
		_designerViewChoice.registerDesignViews(this);

		this.ensureSkin();
		this.layoutParent && this.layoutParent.ensureLayoutSlot(this);
		this.scrollview.appendChild(this.body);
		this.domItem.appendChild(this.contentSlot);
		this.ensureMenuBar();
		this.renderLayoutContent();
		this.layoutSlot && this.layoutSlot.appendChild(this.domItem);
	},
	applyDesignMeta: function() {
		//inportant for authoring
	},
	getMenuItem: function($bind) {
		var bound = this.menuItems[$bind];
		return bound ? bound[0] : null;
	},

	getDataUrl: function() {
		var $parsedUrl = syra_expression.parse(this, this.dataset.$url);
		return $parsedUrl || this.$prototype.$representationUrl;
	},
	fetch: function(options, success) {
		var self = this;
		syra_quality.page_fetch(this, "progress");
		if (options) {
			if (options.isPageLoading) {
				(options.params = options.params || {}).count = _paging.getDefaultItemsPerPage(self);
				delete options.isPageLoading;
			}
			delete options.field;
		}
		options = options || {};
		options.url = options.$url || self.getDataUrl();
		options.page = self;
		options.end = function(response) {
			self.fetchAjax = null;
			if (response.ok) {
				var res;
				if (success) {
					res = success(response.data, response, response.url);
				} else {
					self.startChange();
					self.applyChange(response.data, response, response.url);
					self.endChange();
				}
				syra_quality.page_fetch(self, "success");
				return res;
			} else {
				syra_alert.show(response.error.diagnoses, self);
				syra_quality.page_fetch(self, "error");
			}
		};
		self.fetchAjax = syra_ajax.send(options);
	},
	getBindTitle: function($bind, $field) {
		var $X3Name = $field ? $field.$X3Name : null;
		return $X3Name || $bind;
	},
	getMasterPage: function() {
		return this;
	},
	reloadLayout: function($item, layoutSlot) {
		syra_fields.advancedState.clearPageButton(this);
		this.startChange();
		var prevLayout = this.layoutContent && this.layoutContent.domItem;
		var fields = this.layoutContent.getFields();
		if (fields && fields.length) {
			this.reloadingFields = {};
			for (var ii = 0, jj = fields.length; ii < jj; ii++) {
				var field = fields[ii];
				var $key = field && field.$item && (field.$item.$bind || field.$item.$fieldId);
				if ($key) {
					this.reloadingFields[$key] = field;
				}
			}
		}
		syra_layout.extractItems(this.layoutContent, fields);
		syra_layout.clearContent(this.layoutContent);
		this.$item = $item;
		var autoInsertFields = this.insertNewItems();
		this.ensureSkin();
		this.layoutParent && this.layoutParent.ensureLayoutSlot(this);
		this.renderLayoutContent(layoutSlot);
		this.reloadingFields = null;

		syra_menus.applyChange(this, this.$prototype);
		syra_menus.applyChange(this, this.dataset);
		prevLayout && syra_dom.remove(prevLayout);

		syra_layout.validateLayoutContent(this);
		this.endChange();
		autoInsertFields && autoInsertFields.alert();

		this.garbage && this.garbage.load(this);
	},
	insertNewItems: function(displayDiagnoses, designLevel) {
		// Auto insert fields/sections/blocks in authored page... just for Fusion for now
		this._autoInsertFields && this._autoInsertFields.dispose();
		if (this.isFusionPage === undefined && this.page ? this.page.isFusionPage : this.isFusionPage) {
			(this._autoInsertFields = new _garbage.FieldsAutoInsert()).load(this, designLevel || "article");
		}
		if (this._autoInsertFields && displayDiagnoses) {
			this._autoInsertFields.alert();
		}
		return this._autoInsertFields;
	},
	scrollToItem: function(item, area, isDomItem) {
		if (!isDomItem) {
			var masterPage = item.page.getMasterPage();
			if (masterPage != this) {
				masterPage.scrollToItem(item, area, isDomItem);
				return;
			}
		}
		if (item == this) {
			if (this.scrollview) {
				this.scrollview.scrollTop = this.scrollview.scrollLeft = "0px";
			}
		} else {
			var target;
			area = area || this.scrollview || syra_site.body;
			if (isDomItem) {
				target = item;
			} else {
				var parent = item;
				while (parent) {
					if (parent.tabTitle && !parent.$opened) {
						parent.openTab();
					}
					parent = parent.boxParent;
				}
				if (item && (item.tabTitle || item.domItem)) {
					target = item.tabTitle || item.domItem;
				}
			}
			if (target) {
				var targetRect = syra_dom.getBoundingClientRect(target);
				var areaRect = syra_dom.getBoundingClientRect(area);
				if (targetRect.top < areaRect.top) {
					//target is above area
					area.scrollTop -= ((areaRect.top - targetRect.top) + (areaRect.height > 20 ? 20 : 0));
				} else {
					if (targetRect.top > areaRect.bottom || targetRect.bottom > areaRect.bottom) {
						var diff = areaRect.height - targetRect.height;
						if (diff > 0) {
							area.scrollTop += (targetRect.bottom - areaRect.bottom) + Math.min(20, diff);
						} else {
							//ensure top is visible
							if (targetRect.top > areaRect.bottom) {
								area.scrollTop = 0;
							} else {
								diff = areaRect.bottom - targetRect.top;
								if (diff < 20) {
									area.scrollTop -= (20 - diff);
								}
							}
						}
					}
				}
				if (targetRect.left < areaRect.left) {
					//target is left  area
					area.scrollLeft -= ((areaRect.left - targetRect.left) + (areaRect.width > 20 ? 20 : 0));
				} else {
					if (targetRect.left > areaRect.right || targetRect.right > areaRect.right) {
						var diff = areaRect.width - targetRect.width;
						if (diff > 0) {
							area.scrollLeft += (targetRect.right - areaRect.right) + Math.min(20, diff);
						} else {
							if (targetRect.left > areaRect.right) {
								area.scrollLeft = 0;
							} else {
								diff = areaRect.right - targetRect.left;
								if (diff < 20) {
									area.scrollLeft -= (20 - diff);
								}
							}
						}
					}
				}
			}
		}
	},
	dispose: function() {
		this.fetchAjax && this.fetchAjax.dispose();
		this.breadcrumb && this.breadcrumb.dispose();
		syra_form.remove(this);
		this.isDesigned && syra_site.switchItemDesigner(this, false);
		if (this.bars) { //pagestorage no default load
			while (this.bars.length) {
				this.bars[0].dispose();
			}
		}

		this._autoInsertFields && this._autoInsertFields.dispose();
		this.garbage && this.garbage.dispose();
		this.alertPanel && this.alertPanel.dispose();
		_designerViewChoice.dispose(this);
		syra_pageBuilder.disposePage(this);
		syra_dom.empty(this.body);
		this.loaded = false;
		if (this.menuBar) {
			this.menuBar.slot && syra_dom.remove(this.menuBar.slot);
			this.menuBar.dispose();
		}
		syra_article.dispose(this);
	}
});
