// syracuse-ui/lib/site/bookmarks/bookmarksPage.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

var DesktopPage = require("syracuse-ui/lib/page/desktopPage").DesktopPage;
var BookmarkItemList = require('./bookmarkItemList').BookmarkItemList;
var BookmarkMenuItem = require('./bookmarkMenuItem').BookmarkMenuItem;
var BookmarkMenuBlock = require('./bookmarkMenuBlock').BookmarkMenuBlock;

function BookmarksPage(){
}

exports.BookmarksPage = helpers.defineClass(BookmarksPage, DesktopPage, {
    loadBox: function(){
        DesktopPage.prototype.loadBox.call(this);
    },
    createField: function($field, $item, boxParent, $class){
        switch ($item.$bind) {
            case "items":
                return new BookmarkItemList();
            case "$singleField":
                switch ($item.$variantItemKey) {
                    case "menuItem":
                        return new BookmarkMenuItem();
                    case "menuBlock":
                        return new BookmarkMenuBlock();
                }
                return null;
        }
        return null;
    },
    notifyChangeToServer: function(sendBag, articleSender){
    
    },
    notifyDataChange: function(field, value, notifyServerChange){
        var sendBag = this.ensureSendBag(field.articleParent, field.articleParent);
        sendBag.saveDataChange(field.articleParent, field.getArticle(), field.$item.$bind, field.parentVariantField ? field.parentVariantField.saveVariantValue(value) : value);
        if (notifyServerChange !== false) {
            this.notifyChangeToServer(sendBag, field.articleParent);
        }
        document.site.bookmarksProvider.notifyDataChange(field, value);
        
    },
    ensureReorderAndMenusVisibility: function(item, event, useCssForMenus){
        if (!event) {
            if (useCssForMenus) {
                this.toggleCssOnEnter(item.menusSlot);
            }
            else {
                this.showOnEnter(item.menusSlot);
            }
            this.showOnEnter(item.reorderPicker);
        }
        else {
            if (event) {
                if (useCssForMenus) {
                    this.toggleCssOnEnter(item.menusSlot, event);
                }
                else {
                    this.showOnEnter(item.menusSlot, event);
                }
                this.showOnEnter(item.reorderPicker, event);
            }
        }
    },
    getTreeLevel: function(item){
        var articleParent = item.articleParent;
        var level = 0;
        while (articleParent && articleParent != this) {
            if (articleParent.isMenuBlock) {
                level++
            }
            articleParent = articleParent.articleParent;
        }
        return ("s-level-" + level);
    },
    ensureDesignerOpenerVisibility: function(){
        document.site.enablePageDesign(false);
    },
    dispose: function(){
        DesktopPage.prototype.dispose.call(this);
    }
});
