// syracuse-ui/lib/site/bookmarks/bookmarksManager.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var BookmarksPage = require("./bookmarksPage").BookmarksPage;
var _dialog;

var _barArea;
var _lastCheckedPage;
var _count, _fusionMenuItems, _syracuseMenuItems;

function _refreshItems(items) {
	for (var ii = 0, jj = items.records.length; ii < jj; ii++) {
		var record = items.records[ii];
		if (record.singleField.$variantItemKey == "menuItem") {
			exports.checkNavigationMenuItemBookMark(record.singleField.variantItem);
		} else {
			_refreshItems(record.singleField.variantItem.record.itemsField);
		}

	}
}

exports.refreshNavigationPage = function() {
	if (syra_site.mainPage && syra_site.mainPage.isNavigationPage) {
		var page = syra_site.mainPage;
		if (!page.isAdminMode) {
			var modules = page.boundFields && page.boundFields.modules;
			if (modules) {
				for (var ii = 0, jj = modules.length; ii < jj; ii++) {
					var module = modules[ii];
					for (var mm = 0, kk = module.records.length; mm < kk; mm++) {
						var subModules = module.records[mm].subModulesItem;
						if (subModules) {
							for (var pp = 0, nn = subModules.records.length; pp < nn; pp++) {
								_refreshItems(subModules.records[pp].itemsField);
							}
						}

					}
				}
			}
		}
	}
};

function _refreshManager() {
	if (_dialog) {
		var items = helpers.object.clone(_barArea.dataset.items, true);
		_dialog._content.applyChange({
			items: items
		});
		_dialog.resizeDialog();
	}
}

function _defineMenuItemLinks() {
	return {
		$select: {
			$title: "Add",
			$variants: {
				menuItem: {
					$title: syra_local.bookmarksSelectMenu,
					$url: "/sdata/syracuse/collaboration/syracuse/menuItems?representation=menuItem.$select"
				}
			}
		}
	};
}

function _defineMenuItem() {
	return {
		$type: "application/x-reference",
		$capabilities: "reorder,delete",
		$item: {
			$value: "{title}",
			$key: "{$uuid}",
			$properties: {
				title: {
					$type: "application/x-string"
				}
			}
		}
	};
}

exports.onManagerClick = function() {
	var pageSlot = document.createElement("div");
	pageSlot.className = "s-bookmarks-viewer";
	syra_site.layoutSlot.appendChild(pageSlot);
	_dialog = syra_site.dialogManager.openPage(syra_site, {
		onClose: function(isCanceled, dispose) {
			_dialog = null;
			return true;
		},
		$itemPage: {
			$pageCategoryClass: BookmarksPage,
			layoutSlot: pageSlot,
			$category: "page",
			$facet: "$bookmarks",
			$representation: {
				$prototype: {
					$type: "application/json",
					$title: syra_local.bookmarksTitle,
					$properties: {
						items: {
							$type: "application/x-array",
							$capabilities: "reorder",
							$links: _defineMenuItemLinks(),
							$item: {
								$type: "application/x-variant",
								$variants: {
									menuItem: _defineMenuItem(),
									menuBlock: {
										$type: "application/x-object",
										$capabilities: "append,reorder,delete",
										$item: {
											$value: "{title}",
											$key: "{$uuid}",
											$id: "menuSubblock",
											$properties: {
												title: {
													$type: "application/x-string"
												},
												items: {
													$type: "application/x-array",
													$capabilities: "reorder",
													$links: _defineMenuItemLinks(),
													$item: {
														$type: "application/x-variant",
														$variants: {
															menuItem: _defineMenuItem(),
															menuBlock: {
																$type: "application/x-object",
																$capabilities: "append,reorder,delete",
																$item: {
																	$prototype: "#menuSubblock",
																	$type: "application/x-pointer"
																}
															}
														}

													}

												}
											}
										}
									}
								}
							}
						}
					}
				},
				$article: {
					$layout: {
						$items: [{
							$layout: {
								$items: [{
									$isEditMode: true,
									$bind: "items",
									$skin: "s-bookmarks",
									$layout: {
										$items: [{
											$bind: "menuItem"
										}]
									}
								}]
							}
						}]
					}
				}
			}
		}
	});
	_refreshManager();
};

exports.dispose = function() {
	_resetLinks();
	_barArea = null;
	if (_dialog) {
		_dialog.dispose();
	}
	_dialog = null;
};

function _resetLinks() {
	_lastCheckedPage = null;
	_count = 0;
	_fusionMenuItems = {};
	_syracuseMenuItems = {};
	for (var ii = 0, jj = _barArea.links.length; ii < jj; ii++) {
		delete _barArea.links[ii].menuItem;
	}
	_barArea.links = [];
}

function _addLineLinks(items) {
	if (items) {
		var urlMaker = syra_site.urlMaker;
		var selectedEndpoint = syra_site.userProfile.dataset.selectedEndpoint;
		for (var ii = 0, jj = items.length; ii < jj; ii++) {
			var item = items[ii];
			var link;
			if (item.menuItem) {
				if (item.menuItem.$url) {
					var link = syra_menus.addTextButton(item.menuItem.title || item.menuItem.description || "???", "s-bookmarks-bar-link", "onBarLinkClick");
					if (item.menuItem.$hasSeparator) {
						link.syraHasSeparator = true;
						link.className += " s-separator";
					}
					link.syraSiteObserver = "bookmarksManager";
					link.title = item.menuItem.description;
					item.menuItem.urlSegment = urlMaker.parse(item.menuItem.$url);
					if (urlMaker.replaceSegmentEndpoint(item.menuItem.urlSegment, selectedEndpoint, true)) {
						item.menuItem.$url = urlMaker.build(item.menuItem.urlSegment);
					}
					if (item.menuItem.urlSegment.isFusion) {
						_fusionMenuItems[item.menuItem.urlSegment.fusionParams.fullCode] = item.menuItem;
					} else {
						(_syracuseMenuItems[item.menuItem.urlSegment.uri] = _syracuseMenuItems[item.menuItem.urlSegment.uri] || []).push(item.menuItem);
					}
					link.menuItem = item.menuItem;
					syra_site.history.setHref({
						$url: item.menuItem.$url,
						domItem: link
					});
					_count++;

					_barArea.links.push(_barArea.linksSlot.appendChild(link));
				} else {
					items.splice(ii, 1); //remove old separator
					ii--;
					jj--;
				}
			} else {
				if (item.menuBlock && !_addLineLinks(item.menuBlock.items)) {
					return false;
				}
			}
		}
	}
	return true;
}

exports.onMenuClick = function(menuItem) {
	if (menuItem.$bind && menuItem.$bind.split("-")[0] === "bookmark") {
		exports.onLinkLinkClick(menuItem);
		return false;
	}
	return true;
};

exports.onLinkLinkClick = function(menuItem) {
	var workBook = syra_site.fusionGateway && syra_site.fusionGateway.activatedBook;
	var target;
	if (workBook) {
		if (menuItem.urlSegment.fusionParams) {
			//x3 fct : open via ackcall
			//how to get current field with best way ?
			var sapController = workBook.fusionSite.controller._sapController;
			syra_site.mainPage.externalAdapter.onBlockExRpc({
				field: sapController.getBoundField(sapController._currCtx.ist),
				call: {
					proxy: "EXEFNC1",
					values: [menuItem.urlSegment.fusionParams.functionCode, menuItem.urlSegment.fusionParams.transaction],
					callback: function() {

					}
				}
			});
		}
		return;
		//target = "blank";
	}
	syra_site.history.load(menuItem.urlSegment.$url, target);
};


function _toggleNavigationMenuItemBookMark(navigationMenuItem) {
	syra_menus.setButtonTitle(navigationMenuItem.bookmarkBtn, navigationMenuItem.isBookmark ? syra_local.bookmarksRemove : syra_local.bookmarksAdd);
	syra_site.dom.toggleClass(navigationMenuItem.bookmarkBtn, "s-bookmark-on", navigationMenuItem.isBookmark);
}

function _deleteMenuItem(deletedItem, items) {
	var count = 0;
	items = items || _barArea.dataset.items;
	for (var ii = 0, jj = items.length; ii < jj; ii++) {
		var item = items[ii];
		if (item.menuItem == deletedItem) {
			items.splice(ii, 1);
			return true;
		} else {
			if (item.menuBlock && item.menuBlock.items && _deleteMenuItem(deletedItem, item.menuBlock.items)) {
				return true;
			}
		}
	}
}

function _onNavigationMenuItemPickerClick(navigationMenuItem) {
	if (!navigationMenuItem.isBookmark) {
		if (!_findMenuItem(navigationMenuItem.urlSegment)) {
			navigationMenuItem.isBookmark = true;
			_barArea.dataset.items.push({
				menuItem: {
					$url: navigationMenuItem.urlSegment.$url,
					description: navigationMenuItem.codeMenu.$description || "",
					title: navigationMenuItem.codeMenu.$title || ""
				}
			});
		}
	} else {
		navigationMenuItem.isBookmark = false;
		var menuItem = _findMenuItem(navigationMenuItem.urlSegment);
		if (menuItem) {
			_deleteMenuItem(menuItem);
		}
	}
	_toggleNavigationMenuItemBookMark(navigationMenuItem);
	exports.notifyChangeToServer();
};

function _findMenuItem(urlSegment, items, deleteIfFound) {
	var menuItem;
	if (_count) {
		if (urlSegment.fusionParams) {
			menuItem = _fusionMenuItems[urlSegment.fusionParams.fullCode];
		} else {
			var menuItems = _syracuseMenuItems[urlSegment.uri];
			if (menuItems) {
				for (var ii = menuItems.length - 1; !menuItem && ii >= 0; ii--) {
					var seg2 = menuItems[ii].urlSegment;
					if (!urlSegment.params && !seg2.params) {
						menuItem = menuItems[ii];
					} else {
						if (urlSegment.params && seg2.params) {
							var keys1 = Object.keys(urlSegment.params);
							var keys2 = Object.keys(seg2.params);
							if (keys1.length == keys2.length) {
								for (var mm = keys1.length - 1; mm >= 0; mm--) {
									if (urlSegment.params[keys1[mm]] != seg2.params[keys1[mm]]) {
										break;
									}
								}
								if (mm < 0) {
									menuItem = menuItems[ii];
								}
							}
						}
					}
				}
			}
		}
	}
	return menuItem;
}

exports.checkNavigationMenuItemBookMark = function(navigationMenuItem) {
	navigationMenuItem.urlSegment = syra_site.urlMaker.parse(navigationMenuItem.codeMenu.$url);
	syra_site.urlMaker.replaceSegmentEndpoint(navigationMenuItem.urlSegment, syra_site.userProfile.dataset.selectedEndpoint, true);
	if (_findMenuItem(navigationMenuItem.urlSegment)) {
		navigationMenuItem.isBookmark = true;
		_toggleNavigationMenuItemBookMark(navigationMenuItem);
	} else {
		if (navigationMenuItem.isBookmark) {
			navigationMenuItem.isBookmark = false;
			_toggleNavigationMenuItemBookMark(navigationMenuItem);
		}
	}
};

exports.notifyChangeToServer = function(dataset) {
	var userProfile = syra_site.userProfile;
	_barArea.dataset = dataset || _barArea.dataset;
	syra_controller.sendRequest(userProfile, {
		$location: {
			$url: syra_site.expressionMaker.parse(userProfile, userProfile.$menus.$bookmarks.$url)
		},
		data: {
			$etag: _barArea.dataset.$etag,
			content: _barArea.dataset
		},
		method: "PUT",
		$etag: _barArea.dataset.$etag
	}, function(data, response, requestUrl) {
		if (!userProfile.disposed) {
			exports.refresh();
			_refreshManager();
		}
	});
};

exports.loadBookmarks = function() {
	syra_controller.sendRequest(syra_site.userProfile, {
		$location: {
			$url: syra_site.expressionMaker.parse(syra_site.userProfile, syra_site.userProfile.$menus.$bookmarks.$url)
		},
	}, function(data, response, requestUrl) {
		if (!syra_site.userProfile.disposed) {
			_barArea.dataset = data.content || {};
			_barArea.dataset.items = _barArea.dataset.items || [];
			exports.refresh();
		}
	});
};

exports.refresh = function() {
	syra_site.dom.empty(_barArea.linksSlot);
	syra_site.dom.empty(_barArea.moreItemsSlot);
	_barArea.morePickerSlot.style.visibility = "hidden";
	_resetLinks();
	_addLineLinks(_barArea.dataset.items);
	exports.checkMainPageStatus();
};


exports.onAddClick = function() {
	var title, urlSegment, description;
	var page = syra_site.fusionGateway.getActivatedSheet();
	if (page) {
		var fct = page.$fusionPageMeta.winModel.getFctName();
		if (fct.name && (page.$fusionPageMeta.winModel._mdata.type == 1 || page.$fusionPageMeta.winModel._mdata.type == 2)) {
			if (page.openerUrlSegments.fusionParams.functionCode != fct.name) {
				urlSegment = helpers.object.clone(page.openerUrlSegments, true);
				syra_site.urlMaker.replaceFusionCode(urlSegment, fct.name);
				syra_site.urlMaker.build(urlSegment);
				title = page.getTitle();
			} else {
				urlSegment = page.openerUrlSegments;
				title = fct.title;
			}
			description = syra_local.nvpFunction + ": " + page.openerUrlSegments.fusionParams.functionCode;
		}
	} else {
		page = syra_site.mainPage;
		if (page && page.openerUrlSegments) {
			urlSegment = page.openerUrlSegments;
			title = page.getTitle();
			if (page.$prototype.$representation) {
				description = syra_local.nvpEntity + ": " + page.$prototype.$representation;
				/*if (page.$prototype.$representation != helpers.string.pluralize(page.$prototype.$representation || "")) {
                 description += " (" + syra_local.nvpRepresentation + ": " + page.$prototype.$representation + ")";
                 }*/
			}
		}
	}
	if (urlSegment) {
		var menuItem = _findMenuItem(urlSegment);
		if (menuItem) {
			_deleteMenuItem(menuItem);
		} else {
			menuItem = {
				$url: urlSegment.$url,
				title: title
			};
			if (description) {
				menuItem.description = description;
			}
			_barArea.dataset.items.push({
				menuItem: menuItem
			});
		}
		exports.refreshNavigationPage();
		exports.notifyChangeToServer();

	}
};



exports.checkMainPageStatus = function() {
	var activePage = syra_site.fusionGateway.getActivatedSheet() || syra_site.mainPage;
	if (activePage && !activePage.disposed && activePage != _lastCheckedPage) {
		_lastCheckedPage = activePage;
		var isBookmarkEnabled;
		if (activePage) {
			isBookmarkEnabled = activePage.isBookmarkEnabled;
			if (isBookmarkEnabled === undefined) {
				isBookmarkEnabled = activePage.$item.$isBookmarkEnabled;
			}
			if (isBookmarkEnabled === undefined) {
				isBookmarkEnabled = activePage.$facet != "$edit";
			}
		}
		var isOn = false;
		var functionfusion;
		var menuItem;
		if (activePage) {
			var urlSegment;
			if (activePage.isFusionPage) {
				functionfusion = activePage.$fusionPageMeta && activePage.$fusionPageMeta.winModel.getFctName();
				if (functionfusion) {
					urlSegment = {
						isFusion: true,
						fusionParams: {
							fullCode: functionfusion.name,
							functionCode: functionfusion.name,
							transaction: functionfusion.transaction
						}
					};
					if (functionfusion.transaction) {
						urlSegment.fusionParams.fullCode += "~" + functionfusion.transaction;
					}
					//(urlSegment.fusionParams.transaction == item.menuItem.urlSegment.transaction)) {
				}
			} else {
				urlSegment = syra_site.urlMaker.parse(syra_site.history.getSdataUrl(activePage.openerUrlSegments.$url));
			}
			menuItem = urlSegment && _findMenuItem(urlSegment);
		}
		isOn = !! menuItem;
		for (var ii = 0, jj = _barArea.links.length; ii < jj; ii++) {
			var link = _barArea.links[ii];
			if (link.menuItem) {
				syra_site.dom.toggleClass(link, "s-bookmark-on", link.syraBookMarkOn = menuItem && link.menuItem == menuItem);
				syra_site.dom.disableItem(link, (activePage && activePage.isFusionPage) && !link.menuItem.urlSegment.isFusion);
			}
		}
		_barArea.manageBtn.style.visibility = _count ? "visible" : "hidden";
		var max = syra_site.siteOptions.$item.$maxBookmarks || 25;
		if (_count >= max) {
			isBookmarkEnabled = false;
		}
		syra_site.dom.disableItem(_barArea.addBtn, !isBookmarkEnabled);
		syra_menus.setButtonTitle(_barArea.addBtn, isOn ? syra_local.bookmarksRemove : syra_local.bookmarksAdd);
		syra_site.dom.toggleClass(_barArea.addBtn, "s-bookmark-on", isOn);
		exports.onResize();
	}
};


exports.onResize = function() {
	if (_barArea.linksSlot.clientWidth <= (_barArea.linksSlot.scrollWidth - 2)) {
		for (var ii = _barArea.links.length - 1; ii >= 0; ii--) {
			var link = _barArea.links[ii];
			if (!link.syraIsInMore) {
				link.syraIsInMore = true;
				link.className = "s-bookmarks-bar-link-in-more";
				if (link.syraHasSeparator) {
					link.className += " s-separator";
				}
				if (link.syraIsDisabled) {
					syra_site.dom.disableItem(link, true);
				}
				if (link.syraBookMarkOn) {
					link.className += " s-bookmark-on";
				}
				_barArea.moreItemsSlot.insertBefore(link, _barArea.moreItemsSlot.firstChild);
			}
			if (_barArea.linksSlot.clientWidth >= (_barArea.linksSlot.scrollWidth - 2)) {
				break;
			}
		}
	} else {
		if (_barArea.moreItemsSlot.children.length) {
			for (var ii = 0, jj = _barArea.links.length; ii < jj; ii++) {
				var link = _barArea.links[ii];
				if (link.syraIsInMore) {
					link.className = "s-bookmarks-bar-link";
					if (link.syraIsDisabled) {
						syra_site.dom.disableItem(link, true);
					}
					if (link.syraBookMarkOn) {
						link.className += " s-bookmark-on";
					}
					_barArea.linksSlot.appendChild(link);
					if (_barArea.linksSlot.clientWidth <= (_barArea.linksSlot.scrollWidth - 2)) {
						link.className = "s-bookmarks-bar-link-in-more";
						if (link.syraBookMarkOn) {
							link.className += " s-bookmark-on";
						}
						if (link.syraIsDisabled) {
							syra_site.dom.disableItem(link, true);
						}
						_barArea.moreItemsSlot.insertBefore(link, _barArea.moreItemsSlot.firstChild);
						break;
					} else {
						link.syraIsInMore = false;
					}
					if (link.syraHasSeparator) {
						link.className += " s-separator";
					}
				}
			}
		}
	}
	_barArea.morePickerSlot.style.visibility = _barArea.moreItemsSlot.children.length > 0 ? "visible" : "hidden";
};
exports.onMoreClick = function() {
	if (!_barArea.popupPicker) {
		_barArea.popupPicker = syra_site.dialogManager.openPopup(syra_site, {
			content: _barArea,
			slot: _barArea.moreItemsSlot,
			position: {
				my: "left top",
				at: "left bottom",
				of: $(_barArea.morePickerSlot)
			},
			onClose: function() {
				setTimeout(function() {
					_barArea.popupPicker = null;
				}, 200);
			}
		});
	} else {
		_barArea.popupPicker.close();
	}
};

exports.onBarLinkClick = function(event, btn) {
	exports.onLinkLinkClick(btn.menuItem);
};

exports.onNavigationBookMarkClick = function(event, btn) {
	_onNavigationMenuItemPickerClick(syra_store.get(btn.syraFieldId));
};

exports.load = function(slot) {
	_barArea = {
		dataset: {
			items: []
		},
		links: []
	};

	_barArea.addBtn = syra_menus.addIconButton(syra_local.bookmarksAdd, "s-bookmarks-bar-add", "onAddClick");
	_barArea.addBtn.syraSiteObserver = "bookmarksManager";

	_barArea.manageBtn = syra_menus.addIconButton(syra_local.bookmarksManage, "s-bookmarks-bar-manage", "onManagerClick");
	_barArea.manageBtn.syraSiteObserver = "bookmarksManager";

	var bar = document.createElement("div");
	bar.id = "s-bookmarks-bar";
	slot.appendChild(bar);

	bar.appendChild(_barArea.addBtn);
	bar.appendChild(_barArea.manageBtn);
	_barArea.manageBtn.style.visibility = "hidden";

	_barArea.linksSlot = document.createElement("div");
	_barArea.linksSlot.id = "s-bookmarks-bar-links";
	bar.appendChild(_barArea.linksSlot);

	_barArea.morePickerSlot = document.createElement("div");
	_barArea.morePickerSlot.id = _barArea.morePickerSlot.className = "s-bookmarks-more-picker-slot";
	_barArea.morePickerSlot.style.visibility = "hidden";
	bar.appendChild(_barArea.morePickerSlot);

	var btn = syra_menus.addTextButton(syra_local.bookmarksMore, "s-bookmarks-more-picker", "onMoreClick");
	btn.syraSiteObserver = "bookmarksManager";
	btn.title = syra_local.bookmarksMoreTooltip;
	_barArea.morePickerSlot.appendChild(btn);

	_barArea.moreItemsSlot = document.createElement("div");
	_barArea.moreItemsSlot.className = "s-mn-popus-body";
};
