// syracuse-ui/lib/site/bookmarks/bookmarksPage.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

var DesktopPage = require("syracuse-ui/lib/page/desktopPage").DesktopPage;
var BookmarkItemList = require('./bookmarkItemList').BookmarkItemList;
var BookmarkMenuItem = require('./bookmarkMenuItem').BookmarkMenuItem;
var BookmarkMenuBlock = require('./bookmarkMenuBlock').BookmarkMenuBlock;

function BookmarksPage(){
}

exports.BookmarksPage = helpers.defineClass(BookmarksPage, DesktopPage, {
    loadBox: function(){
        DesktopPage.prototype.loadBox.call(this);
    },
    createField: function($field, $item, boxParent, $class){
        switch ($item.$bind) {
            case "items":
                return new BookmarkItemList();
            case "$singleField":
                switch ($item.$variantItemKey) {
                    case "menuItem":
                        return new BookmarkMenuItem();
                    case "menuBlock":
                        return new BookmarkMenuBlock();
                }
                return null;
        }
        return null;
    },
    notifyChangeToServer: function(sendBag, articleSender){
    
    },
    notifyDataChange: function(field, value, notifyServerChange){
        var sendBag = this.ensureSendBag(field.articleParent, field.articleParent);
        sendBag.saveDataChange(field.articleParent, field.getArticle(), field.$item.$bind, field.parentVariantField ? field.parentVariantField.saveVariantValue(value) : value);
        if (notifyServerChange !== false) {
            this.notifyChangeToServer(sendBag, field.articleParent);
        }
        document.site.bookmarksProvider.notifyDataChange(field, value);
        
    },
    dispose: function(){
        DesktopPage.prototype.dispose.call(this);
    }
});
