// syracuse-ui/lib/site/bookmarks/bookmarksBar.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var BookmarksPage = require("./bookmarksPage").BookmarksPage;
function BookmarksBar(){
}

exports.BookmarksBar = helpers.defineClass(BookmarksBar, null, {
    addPicker: function(css, text, title){
        var link = document.createElement("a");
        link.className = css;
        link.setAttribute("data-s-picker", css);
        link.syraSiteObserver = "bookmarksBar";
        link.setAttribute("href", "#");
        if (text) {
            link.textContent = text;
        }
        if (title) {
            link.title = title;
        }
        return link;
    },
    load: function(slot){
        this.bar = document.createElement("div");
        this.bar.id = "s-bookmarks-bar";
        slot.appendChild(this.bar);
        
        this.bar.appendChild(this.addPicker("s-bookmarks-bar-add", null, document.site.localize.bookmarksAdd));
        
        this.linksSlot = document.createElement("div");
        this.linksSlot.id = "s-bookmarks-bar-links";
        this.bar.appendChild(this.linksSlot);
        
        this.bar.appendChild(this.morePicker = this.addPicker("s-bookmarks-bar-more", document.site.localize.bookmarksMore));
    },
    refresh: function(){
        var site = document.site;
        site.emptyDom(this.linksSlot);
        for (var ii = 0; ii < site.bookmarksProvider.items.length; ii++) {
            var item = site.bookmarksProvider.items[ii];
            if (item.menuItem) {
                var link = this.addPicker("s-bookmarks-bar-link", item.menuItem.title || item.menuItem.description || "???", item.menuItem.description);
                link.syraLink = item.menuItem.$links.$execute_$details.$url;
                this.linksSlot.appendChild(link);
                if (this.linksSlot.clientWidth != this.linksSlot.scrollWidth) {
                    this.linksSlot.removeChild(link);
                    break;
                }
            }
        }
        this.morePicker.style.visibility = site.bookmarksProvider.items.length ? "visible" : "hidden";
    },
    onClickPicker: function(picker, event){
        var $bind = picker.getAttribute("data-s-picker");
        switch ($bind) {
            case "s-bookmarks-bar-link":
                if (picker.syraLink && picker.syraLink !== "") {
                    document.site.bookmarksProvider.navigate(document.site.bookmarksProvider._getItem(picker.syraLink));
                }
                return false;
            case "s-nav-menu-bookmark":
                document.site.bookmarksProvider.toggleNavigationMenuItem(document.site.findField(picker));
                this.refresh();
                return false;
            case "s-bookmarks-bar-add":
                var params = null, url, title;
                var mainPage = document.site.mainPage;
                if (mainPage && mainPage.$isFusionPage) {
                    var fct = mainPage.$urlParts.params.f;
                    params = {};
                    params.f = mainPage.$urlParts.params.f = fct.split("/")[0];
                    if (mainPage.$item.$transaction && mainPage.$item.$transaction !== "") {
                        params.transaction = mainPage.$item.$transaction;
                    }
                }
                if (!params || !params.transaction) {
                    url = "?url=" + encodeURIComponent(document.controller._currentUrl);
                }
                else {
                    var urlParse = helpers.url.parseUrl(document.controller._currentUrl);
                    var af = urlParse.query.f.split("/");
                    af[0] = params.f + "~" + params.transaction;
                    url = "?url=" + encodeURIComponent(urlParse.path + "?f=" + af.join("/"));
                }
                title = mainPage.getTitle();
                document.site.bookmarksProvider.save(url, title, title, params);
                this.refresh();
                return false;
                break;
            case "s-bookmarks-bar-more":
                if (!this.bookmarksPage) {
                    var pageSlot = document.createElement("div");
                    pageSlot.className = "s-bookmarks-viewer";
                    document.site.layoutSlot.appendChild(pageSlot);
                    this.dialog = document.site.openDialog({
                        $dialogMode: "modal",
                        $itemPage: {
                            $pageCategoryClass: BookmarksPage,
                            layoutSlot: pageSlot,
                            $category: "page",
                            $urlParts: {
                                $facet: "$bookmarks"
                            },
                            $representation: {
                                $prototype: {
                                    $type: "application/json",
                                    $title: "Favoris",
                                    $properties: {
                                        items: {
                                            $type: "application/x-array",
                                            $capabilities: "reorder",
                                            $links: {
                                                $select: {
                                                    $title: "Add",
                                                    $variants: {
                                                        menuItem: {
                                                            $title: "Select menu item"
                                                        }
                                                    }
                                                }
                                            },
                                            $item: {
                                                $type: "application/x-variant",
                                                $variants: {
                                                    menuItem: {
                                                        $type: "application/x-reference",
                                                        $capabilities: "reorder,delete",
                                                        $item: {
                                                            $value: "{title}",
                                                            $key: "{$uuid}",
                                                            $properties: {
                                                                title: {
                                                                    $type: "application/x-string"
                                                                }
                                                            }
                                                        }
                                                    },
                                                    menuBlock: {
                                                        $type: "application/x-object",
                                                        $capabilities: "append,reorder,delete",
                                                        $item: {
                                                            $value: "{title}",
                                                            $key: "{$uuid}",
                                                            $id: "menuSubblock",
                                                            $properties: {
                                                                title: {
                                                                    $type: "application/x-string"
                                                                },
                                                                items: {
                                                                    $type: "application/x-array",
                                                                    $capabilities: "reorder",
                                                                    $links: {
                                                                        $select: {
                                                                            $title: "Add",
                                                                            $variants: {
                                                                                menuItem: {
                                                                                    $title: "Select menu item"
                                                                                }
                                                                            }
                                                                        }
                                                                    },
                                                                    $item: {
                                                                        $type: "application/x-variant",
                                                                        $variants: {
                                                                            menuItem: {
                                                                                $type: "application/x-reference",
                                                                                $capabilities: "reorder,delete",
                                                                                $item: {
                                                                                    $value: "{title}",
                                                                                    $key: "{$uuid}",
                                                                                    $properties: {
                                                                                        title: {
                                                                                            $type: "application/x-string"
                                                                                        }
                                                                                    }
                                                                                }
                                                                            },
                                                                            menuBlock: {
                                                                                $type: "application/x-object",
                                                                                $capabilities: "append,reorder,delete",
                                                                                $item: {
                                                                                    $prototype: "#menuSubblock",
                                                                                    $type: "application/x-pointer"
                                                                                }
                                                                            }
                                                                        }
                                                                    
                                                                    }
                                                                
                                                                }
                                                            }
                                                        }
                                                    }
                                                }
                                            }
                                        }
                                    }
                                },
                                $article: {
                                    $layout: {
                                        $items: [{
                                            $category: "section",
                                            $layout: {
                                                $items: [{
                                                    $category: "block",
                                                    $layout: {
                                                        $items: [{
                                                            $isEditMode: true,
                                                            $bind: "items",
                                                            $isQuickDesignerEnabled: false,
                                                            $skin: "s-bookmarks",
                                                            $layout: {
                                                                $items: [{
                                                                    $bind: "menuItem"
                                                                }]
                                                            }
                                                        }]
                                                    }
                                                }]
                                            }
                                        }]
                                    }
                                }
                            }
                        }
                    });
                }
                this.dialog._content.applyChange({
                    items: helpers.object.clone(document.site.bookmarksProvider.items, true)
                });
                return false;
        }
    },
    
    dispose: function(){
        if (this.dialog) {
            this.dialog.dispose();
        }
        this.bookmarksPage = this.bar = this.morePicker = this.linksSlot = null;
    }
});
