// syracuse-ui/lib/field/array/record/selector/buttonRecordSelector.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var RecordSelector = require('./recordSelector').RecordSelector;

function ButtonRecordSelector(){
}

exports.ButtonRecordSelector = helpers.defineClass(ButtonRecordSelector, RecordSelector, {
    load: function(list){
        this.text = list._localize.rsf_select || "Select";
        this.width = (11 * this.text.length) + "px";
        RecordSelector.prototype.load.call(this, list);
    },
    getWidth: function(){
        return this.width;
    },
    appendToTitle: function(slot){
    
    },
    appendToRecord: function(slot){
        slot.className += " s-list-selector-button-slot";
        var btn = document.createElement("a");
        btn.className = "s-list-selector s-list-selector-button";
        btn.setAttribute("title", this.text);
        $(slot.appendChild(btn)).text(this.text);
        return btn;
    },
    onclick: function(options){
        options.isSelected = true;
        options.uuidTarget = this.list._store.findRecordUiid(options.$$selector);
    },
    onSelect: function(options){
        if (this.uuid) {
            this.select(this.uuid, false);
        }
        this.select(options.uuidTarget, options.isSelected);
    }
});
