// syracuse-ui/lib/field/array/record/selector/recordSelector.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function RecordSelector(){
}

exports.RecordSelector = helpers.defineClass(RecordSelector, null, {
    load: function(list){
        this.list = list;
        this.records = {};
        this.bindEvent();
    },
    bindEvent: function(){
        var self = this;
        if (self.list.$item.$selectMode == "row") {
            self.list.$$item.delegate(".s-list-selector-row", "click", function(event){
                self._onSelectRecords(event);
                return false;
            });
        }
        else {
            self.list.$$item.delegate(".s-list-selector,.s-list-selector-all", "click", function(event){
                event.stopPropagation();
                self._onSelectRecords(event);
            });
        }
    },
    clear: function(){
        this.records = {};
        delete this.uuid;
    },
    select: function(uuid, selected){
        var recordArticle = this.list._store._map[uuid];
        recordArticle.highlightSelection(selected);
        if (selected) {
            this.uuid = uuid;
            this.records[uuid] = recordArticle;
        }
        else {
            delete this.uuid;
            delete this.records[uuid];
        }
    },
    _onSelectRecords: function(event){
        document.site.onBeforClick();
        var self = this;
        var options = {
            field: self.list,
            event: event,
            $$selector: $(event.target),
            doEvent: function(){
                self.onSelect(options);
                var articleParent = self.list.getArticleParent();
                if (articleParent.onSelectRecord) {
                    articleParent.onSelectRecord(self.records);
                }
                else {
                    var page = articleParent.getPage();
                    if (page.onSelectRecord) {
                        page.onSelectRecord(self.records);
                    }
                }
                
            }
        };
        self.onclick(options);
        self.list.getPage().externalAdapter.onSelectRecordEvent(options);
    },
    dispose: function(){
        delete this.list;
    }
});
