// syracuse-ui/lib/field/array/record/selector/recordSelector.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function RecordSelector(){
}

exports.RecordSelector = helpers.defineClass(RecordSelector, null, {
    load: function(list){
        this.list = list;
        this.records = {};
    },
    clear: function(){
        this.records = {};
        delete this.uuid;
    },
    select: function(uuid, selected){
        var recordArticle = this.list._store._map[uuid];
        recordArticle.highlightSelection(selected);
        var dataset = recordArticle.ensureDataSet();
        if (selected) {
            this.uuid = uuid;
            this.records[uuid] = recordArticle;
            dataset.$isSelected = true;
        }
        else {
            delete this.uuid;
            delete this.records[uuid];
            delete dataset.$isSelected;
        }
    },
    onSelectMouseEvent: function(event){
        var record = this.list._store.findRecord($(event.target));
        record.onSelectMouseEvent(event);
    },
    onSelectRecords: function(event){
        document.site.onBeforClick();
        var self = this;
        var options = {
            field: self.list,
            event: event,
            $$selector: $(event.target),
            doEvent: function(unselectable){
            	if(unselectable){options.unselectable=unselectable}
                self.onSelect(options);
                if (self.list.articleParent.onSelectRecord) {
                    self.list.articleParent.onSelectRecord(self.records);
                }
                else {
                    if (self.list.page.onSelectRecord) {
                        self.list.page.onSelectRecord(self.records);
                    }
                }
            }
        };
        self.onclick(options);
        self.list.page.externalAdapter.onSelectRecordEvent(options);
    },
    dispose: function(){
        delete this.list;
    }
});
