// syracuse-ui/lib/field/array/record/selector/singleRecordSelector.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var RecordSelector = require('./recordSelector').RecordSelector;

function SingleRecordSelector(){
}

exports.SingleRecordSelector = helpers.defineClass(SingleRecordSelector, RecordSelector, {
    load: function(list){
        this.useColumnSelector = !list.$item.$selectByRowIndex;
        RecordSelector.prototype.load.call(this, list);
    },
    getWidth: function(){
        return 25;
    },
    appendToTitle: function(slot){
    
    },
    appendToRecord: function(slot){
        slot.className += " s-list-selector-slot";
        var input = document.createElement("input");
        input.setAttribute("type", "radio");
        input.className = "s-list-selector";
        return slot.appendChild(input);
    },
    onclick: function(options){
        options.isSelected = this.list.$item.$selectByRowIndex ? true : options.$$selector.is(":checked");
        options.uuidTarget = this.list._store.findRecordUiid(options.$$selector);
    },
    onSelect: function(options){
        if (this.uuid !== undefined) {
            if (this.uuid != options.uuidTarget) {
                this.select(this.uuid, false);
            }
        }
        this.select(options.uuidTarget, options.isSelected);
    }
});
