// @sage/syracuse-react/src/field/poc/string/format.tsx
"use strict";
import * as React from 'react';
import { Icon } from '../../../common/icon/icon';

export interface IFormatType {
    inputType: string,
    icon?: string,
    renderIcon?: () => any,
    renderReadValue: (value: any) => any
}

class Text implements IFormatType {
    get inputType() {
        return "text";
    }
    renderReadValue(value: any) {
        if (value != null && value.toString) {
            value = value.toString();
        }
        return value;
    }
}
class Email implements IFormatType {
    get inputType() {
        return "email";
    }
    get icon() {
        return "mail";
    }
    renderReadValue(value: any) {
        if (value != null && value.length) {
            return <a className="s_link_default" href={"mailto:" + value} rel="nofollow"><Icon name={this.icon}/>{value}</a>;
        }
    }
}
class Phone implements IFormatType {
    get inputType() {
        return "tel";
    }
    get icon() {
        return "phone";
    }
    renderReadValue(value: any) {
        if (value != null && value.length) {
            return  <a className={"s_link_default"} href={"tel:" + value}><Icon name={this.icon}/>{value}</a>;
        }
    }
}
class Url implements IFormatType {
    get inputType() {
        return "url";
    }
    get icon() {
        return "link";
    }
    renderReadValue(value: any) {
        if (value != null && value.length) {
            return  <a className={"s_link_default"} href={value} target="blank"><Icon name={this.icon}/>{value}<Icon name="open_in_new" className="s_open_in_new"/></a>;
        }
    }
}

export let types: any = {
    text: new Text(),
    $email: new Email(),
    $phone: new Phone(),
    $url: new Url()
}
