// syracuse-ui/lib/authoring/field/array/authoringOpener.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Article = require("syracuse-ui/lib/article/article").Article;
var ListAuthoring = require("syracuse-ui/lib/authoring/field/array/listAuthoring").ListAuthoring;

function QuickDesigner(){

}

exports.QuickDesigner = helpers.defineClass(QuickDesigner, null, {
    appendOpener: function(designedField, $$opener){
        var self = this;
        self.designedField = designedField;
        if (!$$opener) {
            self.isQuickDesigner = true;
            self._opener = document.createElement("a");
            self._opener.className = "s-list-quick-designer";
            self._opener.setAttribute("href", "#");
            self.designedField.$$opener = $(self._opener);
            var dom = document.createElement("div");
            dom.className = "s-list-quick-designer-cell";
            dom.appendChild(self._opener);
            $(dom).prependTo($(designedField._topbar));
        }
        else {
            self.designedField.$$opener = $$opener;
        }
        designedField.$$opener.bind("click", function(){
            document.site.onBeforClick();
            self.toggleQuick();
            return false;
        });
    },
    toggleQuick: function(){
        var self = this;
        if (!self.popup) {
            (self.authoring = new ListAuthoring()).loadBox(self);
            self.popup = self.designedField.boxParent.openDialog({
                $dialogMode: "popup",
                page: self.authoring,
                $$dialog: self.authoring.$$container,
                onClose: function(){
                    if (self.authoring.onClose(this)) {
                        self.popup = null;
                        document.controller.disposeObject(self.authoring);
                        delete self.authoring;
                        return true;
                    }
                    return false;
                },
                onValidate: function(){
                    return self.authoring.onValidate(this);
                }
            });
        }
        else {
            if (self.popup) {
                self.popup.close();
            }
        }
    },
    dispose: function(){
        if (this.designedField) {
            if (this.designedField.$$opener) {
                this.designedField.$$opener.unbind();
                this.designedField.$$opener = null;
            }
            this.designedField = null;
        }
    }
});

function WidgetAuthoring(){

}

exports.WidgetAuthoring = helpers.defineClass(WidgetAuthoring, Article, {
    loadBox: function(initData){
        if (!this.$$item) {
            var dom = document.createElement("a");
            dom.className = "s-list-main-designer";
            this.$$item = $(dom);
            dom = document.createElement("div");
            dom.className = "s-list-main-designer-icon s-top-link-aw-open";
            this.$$item.append(dom);
            
            var domTitle = document.createElement("div");
            domTitle.className = "s-list-main-designer-title";
            this.$$item[0].appendChild(domTitle).textContent = "Design the current list";
            
            
            (this.quickDesigner = new QuickDesigner()).appendOpener(this.designedField, this.$$item.appendTo(this.$$container));
        }
    },
    dispose: function(){
        delete this.designedField;
        Article.prototype.dispose.call(this);
    }
});
