// syracuse-ui/lib/field/binary/imageFieldAuthoring.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Article = require("syracuse-ui/lib/article/article").Article;

function ImageFieldAuthoring(){

}

exports.ImageFieldAuthoring = helpers.defineClass(ImageFieldAuthoring, Article, {
    applyChange: function(newData){
        var metaData = {
            $properties: {}
        };
        if (newData) {
            Object.keys(newData).forEach(function($prop){
                var value = newData[$prop];
                if ($prop == "$isAutoSize") {
                    metaData.$properties.$imageHeight = {
                        $isDisabled: value
                    };
                }
                metaData[$prop] = value;
            });
        }
        Article.prototype.applyChange.call(this, metaData);
    },
    loadBox: function(initData){
        this.$prototype = {
            "$properties": {
                $imageWidth: {
                    $type: "application/x-string",
                    $title: this.designedField._localize.if_imageFieldWidth
                },
                $imageHeight: {
                    $type: "application/x-string",
                    $title: this.designedField._localize.if_imageFieldHeight
                },
                $isAutoSize: {
                    $type: "application/x-boolean",
                    $title: this.designedField._localize.if_imageFieldWidthIsAutoSize
                }
            }
        };
        this.$item = {
            $category: "section",
            $layout: {
                $items: [{
                    $bind: "$isAutoSize",
                    $isEditMode: true,
                    $labelWidth: "auto",
                    $skin: "s-aw-field"
                }, {
                    $bind: "$imageHeight",
                    $isEditMode: true,
                    $labelWidth: "auto",
                    $skin: "s-aw-field"
                }]
            }
        };
        Article.prototype.loadBox.call(this, initData);
    },
    dispose: function(){
        Article.prototype.dispose.call(this);
    }
});
