// @sage/syracuse-react/src/test/testCases.ts
"use strict";
export interface IItem {
    url: string,
    title: string
}
export interface IGroup {
    title: string,
    records: IItem[]
}

export const groups: IGroup[] = [{
    title: "Array",
    records: [{
        url: "?representation=s-uitest.$test&name=array_array&category=grid",
        title: "Array Grid"
    }, {
        url: "?representation=s-uitest.$test&name=array_array&category=gridcards",
        title: "Array Grid - Cards"
    }, {
        url: "?representation=s-uitest.$test&name=array_array&category=cards",
        title: "Array Cards"
    }, {
        url: "?representation=s-uitest.$test&name=array_array&category=selection",
        title: "Array Selection"
    }, {
        url: "?representation=s-uitest.$test&name=array_graph",
        title: "Array Graph"
    }, {
        url: "?representation=s-uitest.$test&name=array_tree",
        title: "Array Tree "
    }, {
        url: "?representation=s-uitest.$test&name=array_single",
        title: "Array Single"
    }]

}, {
    title: "Fields",
    records: [{
        title: "UI React Field string",
        url: "?representation=fields.$__tests__&type=string"
    },{
        title: "UI React Field choice",
        url: "?representation=fields.$__tests__&type=choice"
    },{
        title: "UI React Field boolean",
        url: "?representation=fields.$__tests__&type=boolean"
    }, {
        url: "?representation=s-uitest.$test&name=field_field&category=simple",
        title: "Fields"
    },
    {
        url: "?representation=s-uitest.$test&name=field_field&category=xmlJSON",
        title: "Fields Rich Data"
    },
    {
        url: "?representation=s-uitest.$test&name=field_choice&category=choice",
        title: "Fields Choice"
    }, {
        url: "?representation=s-uitest.$test&name=field_field&category=binary",
        title: "Fields Binary"
    }, {
        url: "?representation=s-uitest.$test&name=field_field&category=editor",
        title: "Fields Editor"
    }, {
        url: "?representation=s-uitest.$test&name=field_child",
        title: "Fields Child Object"
    }, {
        url: "?representation=s-uitest.$test&name=field_field&category=icon",
        title: "Fields Icon"
    }]
}, {
    title: "Navigation",
    records: [{
        title: "UI React Breadcrumb",
        url: "?representation=breadcrumb.$__tests__"
    }, {
        title: "UI React Bookmark",
        url: "?representation=bookmark.$__tests__"
    }, {
        title: "UI React Sitemap",
        url: "?representation=sitemap.$__tests__"
    }]
}, {
    title: "Misc",
    records: [{
        title: "UI React Scroll",
        url: "?representation=scroll.$__tests__"
    }, {
        title: "UI React Resizer",
        url: "?representation=resizer.$__tests__"
    }]
}, {
    title: "Misc",
    records: [{
        title: "UI React Alert",
        url: "?representation=alert.$__tests__"
    }, {
        url: "?representation=s-uitest.$test&name=misc_office",
        title: "office",
    }, {
        url: "?representation=over.$__tests__",
        title: "REact - misc Over"
    }, {
        url: "?representation=s-uitest.$test&name=process_process",
        title: "Process"
    }, {
        title: "UI React Tracker",
        url: "?representation=tracker.$__tests__"
    }, {
        url: "?representation=s-uitest.$test&name=misc_parameters",
        title: "Dynamic Parameters"
    }]
}, {
    title: "Complex",
    records: [{
        url: "?representation=s-uitest.$test&name=complex_arrayComplexFields",
        title: "Array complex fields"
    }]
}, {
    title: "extends",
    records: [{
        url: "?representation=s-uitest.$test&name=extend_landingCalendar&view=calendar",
        title: "Landing Page Calendar",
    }, {
        title: "ZCC5 Calendar",
        url: "/trans/x3/erp/superv/$sessions?f=ZCC5"
    }, {
        url: "?representation=s-uitest.$test&name=ui-sql-designer_designer",
        title: "ui-sql-designer",
    }, {
        url: "?representation=s-uitest.$test&name=ui-d3-graph_gantt",
        title: "ui-d3-graph-gantt",
    }]

}, {
    title: "Convergence",
    records: [{
        url: "?representation=s-uitest.$test&name=tools_convergenceCleaner",
        title: "Convergence Cleaner",
    }, {
        title: "X3 link --> AQFA : Fields",
        url: "/trans/x3/erp/superv/$sessions?f=AQFA"
    }, {
        title: "X3 link --> AQFB : Array",
        url: "/trans/x3/erp/superv/$sessions?f=AQFB"
    }, {
        title: "X3 link --> AQFC : Picking",
        url: "/trans/x3/erp/superv/$sessions?f=AQFC"
    }, {
        title: "X3 link --> AQFE : windows",
        url: "/trans/x3/erp/superv/$sessions?f=AQFE"
    }, {
        title: "X3 link --> AQFD : Left list",
        url: "/trans/x3/erp/superv/$sessions?f=AQFD"
    }, {
        title: "X3 link --> AQFF : Flash",
        url: "/trans/x3/erp/superv/$sessions?f=AQFF"
    }, {
        title: "X3 link --> AQFG : ORG chart",
        url: "/trans/x3/erp/superv/$sessions?f=AQFG"
    }]
}, {
    title: "Json",
    records: [{
        title: "JsonLint",
        url: "http://jsonlint.com/"
    }, {
        title: "Jsonformatter Curiousconcept",
        url: "http://jsonformatter.curiousconcept.com/"
    }]
}, {
    title: "Notification",
    records: [{
        title: "test Notification",
        url: "?representation=s-uitest.$test&name=notifications_testNotif"
    }]
}
];
