// syracuse-ui/lib/field/array/gridBuilder/tableHelper.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function _init(tb, list) {
	tb.headSlot = document.createElement("div");
	tb.headTable = document.createElement("table");
	tb.headColgroup = document.createElement("colgroup");
	tb.titleRow = document.createElement("tr");
	if (list.$item.$isTitleRowHidden) {
		tb.titleRow.style.display = "none";
	}

	tb.bodySlot = document.createElement("div");
	tb.bodyTable = document.createElement("table");
	tb.bodyColgroup = document.createElement("colgroup");
	tb.bodyBody = document.createElement("tbody");

	tb.headSlot.className = list.$skin + "-slot-head";
	tb.headTable.className = list.$skin + "-table-head";
	tb.headTable.setAttribute("cellspacing", "0");
	tb.headTable.setAttribute("cellpadding", "0");
	tb.headTable.appendChild(tb.headColgroup);
	tb.headTable.appendChild(tb.titleRow);
	tb.headSlot.appendChild(tb.headTable);

	tb.bodySlot.className = list.$skin + "-slot-body";
	tb.bodyTable.className = list.$skin + "-table-body";
	tb.bodyTable.setAttribute("cellspacing", "0");
	tb.bodyTable.setAttribute("cellpadding", "0");
	tb.bodyTable.appendChild(tb.bodyColgroup);
	tb.bodyTable.appendChild(tb.bodyBody);
	tb.bodySlot.appendChild(tb.bodyTable);
}

function _addHScroller(tb, list) {
	var hscroller = tb.hscroller = document.createElement("div");
	hscroller.syraIsMouseEventDisabled = true;
	hscroller.className = list.$skin + "-h-scroll";
	hscroller.syraNoDDSpot = true;
	tb.hscrollBody = hscroller.appendChild(document.createElement("div"));
	tb.hscrollBody.className = list.$skin + "-h-scroll-body";
}


function Table() {}

helpers.defineClass(Table, null, {
	load: function(list, isScrollable) {
		_init(this, list);
		isScrollable && _addHScroller(this, list);
	},
	addCol: function() {
		return {
			slot: this.titleRow,
			titleCell: this.titleRow.appendChild(document.createElement("th")),
			groupBody: this.bodyColgroup.appendChild(document.createElement("col")),
			groupHead: this.headColgroup.appendChild(document.createElement("col"))
		};
	},
	dispose: function() {
		this.headSlot = this.headTable = this.headColgroup = this.bodySlot = this.bodyTable = this.bodyColgroup = this.bodyBody = null;
		this.hscroller.className = this.hscrollBody = this.cols = null;
	}
});


exports.create = function(list, isScrollable) {
	var tb = new Table();
	tb.load(list, isScrollable);
	return tb;
};
