// syracuse-ui/lib/site/resize/widgetResizerDDAgent.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function WidgetResizerDDAgent() {

}

exports.WidgetResizerDDAgent = helpers.defineClass(WidgetResizerDDAgent, null, {
	start: function(dropItem) {
		this.resizer = dropItem;
		syra_dd.ddAgent = this;
		this.prevDocumentCursor = syra_site.layoutSlot.style.cursor;
		syra_site.layoutSlot.style.cursor = this.resizer.cursor;
		if (this.resizer.options.onResize) {
			this.resizer.options.onResize(this.resizer);
		}
	},
	onDragMouseMove: function(target, event) {
		var minHeight = this.resizer.options.minHeight;
		var minWidth = this.resizer.options.minWidth;
		var width, height, isResized;
		var style = this.resizer.options.slot.style;
		var pageX = event.pageX;
		var pageY = event.pageY;
		if (this.resizer.isLeft) {
			width = this.resizer.right - pageX;
			if (!this.resizer.maxWidth || width < this.resizer.maxWidth) {
				if (!minWidth || (width >= minWidth) || (pageX < (this.resizer.left - 5))) {
					isResized = true;
					style.left = pageX + "px";
					style.width = width + "px";
				}
			}
		} else {
			if (this.resizer.isRight) {
				width = pageX - this.resizer.left;
				if (!this.resizer.maxWidth || width < this.resizer.maxWidth) {
					if (!minWidth || (width >= minWidth) || (pageX > (this.resizer.right - 5))) {
						isResized = true;
						style.width = width + "px";
					}
				}
			}
		}
		if (this.resizer.isTop) {
			height = this.resizer.bottom - pageY;
			if (!this.resizer.maxHeight || height < this.resizer.maxHeight) {
				if (!minHeight || (height >= minHeight) || (pageY < (this.resizer.top - 5))) {
					isResized = true;
					style.top = pageY + "px";
					style.height = height + "px";
				}
			}
		} else {
			if (this.resizer.isBottom) {
				height = pageY - this.resizer.top;
				if (!this.resizer.maxHeight || height < this.resizer.maxHeight) {
					if (!minHeight || (height >= minHeight) || (pageY > (this.resizer.bottom - 5))) {
						isResized = true;
						style.height = height + "px";
					}
				}
			}
		}
		if (isResized && this.resizer.options.onResize) {
			this.resizer.options.onResize(this.resizer, true);
		}
	},
	onDragMouseUp: function(target, event) {},
	dispose: function() {
		syra_site.layoutSlot.style.cursor = this.prevDocumentCursor;
		if (this.resizer) {
			if (this.resizer.options.onResize) {
				this.resizer.options.onResize(this.resizer);
			}
		}
	}
});
