// syracuse-ui/lib/site/resize/widgetMoveDDAgent.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function WidgetMoveDDAgent() {

}

exports.WidgetMoveDDAgent = helpers.defineClass(WidgetMoveDDAgent, null, {
	start: function(dropItem) {
		this.resizer = dropItem;
		document.site.ddManager.ddAgent = this;
		document.site.getBoundingClientRect(this.resizer.options.slot, this);
		this.boundary = document.site.body.getBoundingClientRect();
	},
	onDragMouseMove: function(target, event) {
		if (this.startX == undefined) {
			this.startX = event.pageX;
			this.startY = event.pageY;
		}
		var deltaY = event.pageY - this.startY;
		var deltaX = event.pageX - this.startX;

		var newTop = this.top + deltaY;
		var newLeft = this.left + deltaX;

		if ((this.boundary.top < (newTop - 10)) && (this.boundary.left < (newLeft - 10)) && (this.boundary.bottom > (newTop + this.height + 10)) && (this.boundary.right > (newLeft + this.width + 10))) {
			this.resizer.options.slot.style.top = newTop + "px";
			this.resizer.options.slot.style.left = newLeft + "px";
			if (this.resizer.options.source) {
				this.resizer.options.source.draggedPosition = {
					top: newTop,
					left: newLeft
				};
			}
		}
		event.preventDefault();
		return false;
	},
	onDragMouseUp: function(target, event) {},
	dispose: function() {
		this.resizer = null;
	}
});
