// syracuse-ui/lib/site/resize/widgetMoveDDAgent.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function WidgetMoveDDAgent() {

}

exports.WidgetMoveDDAgent = helpers.defineClass(WidgetMoveDDAgent, null, {
	start: function(dropItem) {
		this.resizer = dropItem;
		syra_dd.ddAgent = this;
		syra_site.dom.getBoundingClientRect(this.resizer.options.slot, this);
		var bodyRect = syra_site.getBodySize();
		this.maxTop = bodyRect.bottom - this.height;
		this.maxLeft = bodyRect.right - this.width;
	},
	onDragMouseMove: function(target, event) {
		if (this.startX == undefined) {
			this.startX = event.pageX;
			this.startY = event.pageY;
		}
		var deltaY = event.pageY - this.startY;
		var deltaX = event.pageX - this.startX;

		var newTop = this.top + deltaY;
		var newLeft = this.left + deltaX;

		newTop = Math.max(newTop, 0);
		newLeft = Math.max(newLeft, 0);
		newTop = Math.min(newTop, this.maxTop);
		newLeft = Math.min(newLeft, this.maxLeft);

		this.resizer.options.slot.style.top = newTop + "px";
		this.resizer.options.slot.style.left = newLeft + "px";
		this.resizer.options.source.isUserPosition = true;
		event.preventDefault();
		return false;
	},
	onDragMouseUp: function(target, event) {},
	dispose: function() {
		this.resizer = null;
	}
});
