// tedious/lib/token/returnvalue-token-parser.js
// s2.2.7.16

'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _metadataParser = require('../metadata-parser');

var _metadataParser2 = _interopRequireDefault(_metadataParser);

var _valueParser = require('../value-parser');

var _valueParser2 = _interopRequireDefault(_valueParser);

exports['default'] = _regeneratorRuntime.mark(function callee$0$0(parser, colMetadata, options) {
  var paramOrdinal, paramName, metadata, value;
  return _regeneratorRuntime.wrap(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        context$1$0.next = 2;
        return parser.readUInt16LE();

      case 2:
        paramOrdinal = context$1$0.sent;
        return context$1$0.delegateYield(parser.readBVarChar(), 't0', 4);

      case 4:
        paramName = context$1$0.t0;

        if (paramName.charAt(0) === '@') {
          paramName = paramName.slice(1);
        }

        context$1$0.next = 8;
        return parser.readUInt8();

      case 8:
        return context$1$0.delegateYield((0, _metadataParser2['default'])(parser, options), 't1', 9);

      case 9:
        metadata = context$1$0.t1;
        return context$1$0.delegateYield((0, _valueParser2['default'])(parser, metadata, options), 't2', 11);

      case 11:
        value = context$1$0.t2;
        return context$1$0.abrupt('return', {
          name: 'RETURNVALUE',
          event: 'returnValue',
          paramOrdinal: paramOrdinal,
          paramName: paramName,
          metadata: metadata,
          value: value
        });

      case 13:
      case 'end':
        return context$1$0.stop();
    }
  }, callee$0$0, this);
});
module.exports = exports['default'];
// status
