// syracuse-ui/lib/field/helpers/advancedState.js
"use strict";

function _ensureFieldVisibility(field) {
	if (field.variantItem) {
		_ensureFieldVisibility(field.variantItem);
	} else {
		syra_fields.showField(field, !field.$isHidden && (!field.boxParent.hasAdvancedFields || field.boxParent.advancedFieldsVisibled));
	}
}

function _showFields(box, show) {
	box.advancedFieldsVisibled = show;
	if (box.isDesktopPage) {
		if (!box.isVignettePage) {
			var title = box.advancedFieldsVisibled ? syra_local.box_advanced_hide : syra_local.box_advanced;
			var icon = box.advancedFieldsVisibled ? "advanced_hide" : "advanced_show";
			syra_button.setText(box.advancedFieldsBtn, title, icon);
		}
	} else {
		if (box.vignette) {
			_showFields(box.vignette, box.vignette.advancedFieldsVisibled = box.advancedFieldsVisibled);
		}
	}
	if (box.layoutContent) {
		var children = box.layoutContent.getItems();
		for (var ii = 0, jj = children.length; ii < jj; ii++) {
			var child = children[ii];
			if (child.isSection || child.vignette) {
				_showFields(child, box.advancedFieldsVisibled);
			} else {
				if (child.$item.$isAdvanced) {
					_ensureFieldVisibility(child);
				}
			}
		}
	}
	box.isDesktopPage && syra_layout.ensureArticleVisibility(box);
}

function _showButton(page, show, designing) {
	if (show && page.hasAdvancedFields) {
		return; //adance bttuons are already visible
	}
	if (page.isDesktopPage) {
		page.hasAdvancedFields = show;
		if (page.isVignettePage) {
			_showButton(page.vignetteField, show, designing);
		} else {
			var title = page.advancedFieldsVisibled ? syra_local.box_advanced_hide : syra_local.box_advanced;
			var icon = page.advancedFieldsVisibled ? "advanced_hide" : "advanced_show";
			if (!page.advancedFieldsBtn) {
				page.advancedFieldsBtn = syra_button.add({
					parent: page,
					slot: page.header,
					text: title,
					css: "s-page-advanced-fields-link",
					fontIcon: icon,
					click: function() {
						_showFields(this.parent, !this.parent.advancedFieldsVisibled);
					}
				});
			} else {
				syra_button.setText(page.advancedFieldsBtn, title);
			}
			syra_button.setText(page.advancedFieldsBtn, title, icon);
			syra_button.hide(page.advancedFieldsBtn, !page.hasAdvancedFields);
		}
	} else {
		if (!show) {
			var items = page.layoutContent.getItems();
			for (var ii = 0, jj = items.length; ii < jj; ii++) {
				if (items[ii].$item.$isAdvanced) {
					return; //has advanced fields, can not hide page 
				}
			}
		}
		page.hasAdvancedFields = show;
		page.boxParent && _showButton(page.boxParent, show, designing);
	}
};

exports.setState = function(field, $isAdvanced, designing) {
	if ($isAdvanced === undefined) {
		if (field.$field.$isAdvanced !== undefined || field.$item.$isAdvanced !== undefined) {
			if (field.$item.$isAdvanced === undefined) {
				field.$item.$isAdvanced = field.$field.$isAdvanced;
			}
			exports.setState(field, field.$item.$isAdvanced);
		}
	} else {
		if (field.variantItem) {
			exports.setState(field.variantItem, $isAdvanced, designing);
		} else {
			if (designing) {
				if ($isAdvanced) {
					field.$item.$isAdvanced = true;
				} else {
					delete field.$item.$isAdvanced;
				}

			}
			_showButton(field.boxParent, $isAdvanced, designing);
			_ensureFieldVisibility(field);
		}
	}
};

exports.toggleAllFields = function(page, show) {
	if (page.advancedFieldsBtn) {
		_showFields(page, show === undefined ? page.advancedFieldsVisibled : show);
	}
};


exports.clearPageButton = function(page) {
	if (page.advancedFieldsBtn) {
		delete page.advancedFieldsVisibled;
		delete page.hasAdvancedFields;
		syra_button.remove(page.advancedFieldsBtn);
		page.advancedFieldsBtn = null;
	}
};
