// syracuse-ui/lib/field/helpers/embeddedHelper.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

exports.injectTestLinks = function(page) {
	if (page.openerUrlSegments && page.openerUrlSegments.representationRoot == "employee") {
		switch (page.openerUrlSegments.$facet) {
			case "$edit":
			case "$details":
				var $links = page.$prototype.$links = page.$prototype.$links || {};
				$links.inlineEmployees = {
					$method: "GET",
					$target: "embedded",
					$title: "employees",
					$type: "application/json;vnd.sage=syracuse",
					$url: "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&where=(firstName%20like%20%27%25{firstName}%25%27)"
				};
				/*$links.inlineSage = {
             $method: "GET",
             $target: "embedded",
             $title: "Sage web",
             $type: "html",
             $url: "http://www.sage.com"
             };*/
				break;
			case "$query":
				page.$prototype.$properties.$resources.$capabilities = "quickedit";
				break;
		}
	}
};

exports.addEmbeddedLink = function(article, $bind, $menu) {
	if (article.isDesktopPage) {
		var $properties = article.$prototype.$properties;
		if ($properties) {
			var $vignette = $properties[$bind];
			$menu = helpers.object.clone($menu, true);
			if ($menu.$url) {
				//$menu.$url = syra_site.expressionMaker.parseUrl(article, $menu.$url);
				var seg = syra_site.urlMaker.parse($menu.$url, true);
				if (!seg.isSyracuse) {
					$menu.$type = "html";
				}
			}
			if (!$vignette) {
				$vignette = $properties[$bind] = {
					$type: "application/x-vignette",
					$title: $menu.$title,
					$location: $menu
				};
			} else {
				syra_site.deltaManager.applyObjectDelta(article.page, $vignette.$location, $menu);
			}
			article.layoutContent.loadChildItem(null, {
				$bind: $bind
			});
			return true;
		}
	}
};


exports.hasMasterChildRelation = function(field) {
	if (field.$field.$location && field.$field.$location.$target == "embedded") {
		var keys = field.$field.$location.$url.match(/(\{.*?\})/g);
		field.masterKeys = {};
		if (keys) {
			for (var ii = 0, jj = keys.length; ii < jj; ii++) {
				var key = keys[ii];
				keys[ii] = key.substr(1, key.length - 2);
				field.masterKeys[keys[ii]] = undefined;
			}
			(field.articleParent.masterChildRelations = field.articleParent.masterChildRelations || []).push(field);
			return true;
		}
	}
	return false;
};


function _refreshChildRelation(vignetteField) {
	vignetteField.$url = syra_site.expressionMaker.parseUrl(vignetteField.articleParent, vignetteField.$field.$location.$url, vignetteField.articleParent.dataset, vignetteField.masterKeys);
	if (!vignetteField.vignette) {
		vignetteField.loadContentRepresentation();
	} else {
		vignetteField.renderLayoutContent(vignetteField.$url);
	}
}

exports.applyChangeToDetailRelations = function(article) {
	if (article.masterChildRelations && article.masterChildRelations.length) {
		for (var ii = 0, jj = article.masterChildRelations.length; ii < jj; ii++) {
			var relation = article.masterChildRelations[ii];
			var refresh = false;
			var keys = Object.keys(relation.masterKeys);
			for (var mm = 0, kk = keys.length; mm < kk; mm++) {
				var key = keys[mm];
				if (article.dataset[key] != relation.masterKeys[key]) {
					relation.masterKeys[key] = article.dataset[key];
					refresh = true;
				}
			}
			refresh && _refreshChildRelation(relation);
		}
	}
};

exports.unregisterMasterChildRelation = function(field) {
	var relations = field.articleParent && field.articleParent.masterChildRelations;
	if (relations) {
		var found = relations.indexOf(field);
		relations.splice(found, 1);
	}
};
