// syracuse-ui/lib/field/helpers/menusHelper.js
"use strict";

function _isPopupMenuItem(field, $bind, $link) {
	if (field.isPopupMenuItem && !field.isPopupMenuItem($bind, $link)) {
		return false;
	}
	switch ($bind) {
		case "$suggest":
			field._isAutoComplete = true;
			return false;
		case "$details":
			if (!field.$isDetailLinkDisabled) {
				return false;
			}
			break;
		case "$localize":
			if (field.$item.$useLocalizePicker) {
				$link.$isHidden = true;
			}
			break;
		case "$tunnel":
			if (field.$item.$inplace) {
				$link.$title = syra_local.fieldTunnel;
			}
			return field.$item.$inplace && !field.$item.$isFilterMode;
		case "$lookup":
		case "$select":
			return false;
	}
	return true;
}

function _applyMenusChange(field, $deltaMenu, $isAction) {
	var binds = Object.keys($deltaMenu);
	for (var ii = 0, jj = binds.length; ii < jj; ii++) {
		var $bind = binds[ii];
		var $menu = $deltaMenu[$bind];
		if (_isPopupMenuItem(field, $bind, $menu)) {
			if ($menu === null) {
				delete field.$menus && field.$menus[$bind];
			} else {
				if (!field.$menus) {
					field.$menus = {};
				}
				if (field.$menus[$bind]) {
					syra_site.deltaManager.applyObjectDelta(field.page, field.$menus[$bind], $menu);
				} else {
					$menu.$isAction = $isAction;
					$menu.$bind = $bind;
					field.$menus[$bind] = $menu;
				}
			}
		}
	}
}


exports.onMenuPickerClick = function(field, picker) {
	if (!field.authoringNode) {
		field.page.externalAdapter.onFieldClickPicker({
			field: field,
			pickerType: "menus",
			doEvent: function(options) {
				var $item = {
					$isOwner: true,
					$noText: true,
					$title: syra_local.fieldActions,
					$clientId: field.id + "-menu-picker",
					$skin: field.$skin + "-menus s-field-context-menus",
					$itemSkin: "s-list-menus-record-menus-link",
					$category: "menus",
					$layout: {
						$items: []
					}
				};
				if (field.$rootLinks && field.$rootLinks.$links && field.$rootLinks.$links.$localize) {
					$item.$layout.$items.push({
						$category: "link",
						$bind: $item.$clientId + "-$localize",
						$sourceBind: "$localize",
						$icon: {
							$mode: "iconText",
							$path: "field/s-field-"
						}
					});
				}
				if (field.$item.$inplace && !field.$item.$isFilterMode) {
					$item.$layout.$items.push({
						$category: "link",
						$bind: $item.$clientId + "-$tunnel",
						$sourceBind: "$tunnel",
						$icon: {
							$mode: "iconText",
							$path: "field/s-field-"
						}
					});
				}
				field.menusBox = field.page.loadNewItem(field.input ? field.fieldValue : field._dataValue, $item, field.articleParent);
				field.menusBox.contextField = field;
				field.menusBox.ensurePrivateMenus(field.$menus);
			}
		});
	}
};
