// syracuse-collaboration/lib/representations/mobile/mobileDashboard.js
"use strict";
exports.representation = {
	$entityName: "mobileDashboard",
	$facets: {
		$details: {
			$layout: {
				$layoutType: "row",
				$items: [{
					$items: [{
						$category: "section",
						$title: "{@InformationsSectionTitle}",
						$layout: {
							$items: [{
								$bind: "title"
							}, {
								$bind: "dashboardName"
							}, {
								$bind: "description"
							}]
						}
					}, {
						$category: "section",
						$title: "{@ContentSectionTitle}",
						$layout: {
							$items: [{
								$bind: "vignettes",
								$layout: {
									$items: [{
										$bind: "gadget"
									}, {
										$bind: "displayStyle"
									}]
								}
							}]
						}
					}]
				}]

			}
		},
		$edit: {
			$copy: "$details"
		},
		$query: {
			$layout: {
				$items: [{
					$category: "section",
					$layout: {
						$items: [{
							$category: "section",
							$layout: {
								$items: [{
									$bind: "$resources",
									$layout: {
										$items: [{
											$bind: "dashboardName",
										}, {
											$bind: "title",
										}, {
											$bind: "description"
										}]
									}
								}]
							}
						}]
					}
				}]
			}
		},
		$lookup: {
			$copy: "$query"
		},
		$select: {
			$copy: "$query"
		}
	}
};
