// syracuse-collaboration/lib/representations/mobile/mobileGadget.js
"use strict";
exports.representation = {
	$entityName: "mobileGadget",
	$facets: {
		$details: {
			$layout: {
				$layoutType: "row",
				$items: [{
					$items: [{
						$category: "section",
						$title: "{@InformationsSectionTitle}",
						$layout: {
							$items: [{
								$bind: "code"
							}, {
								$bind: "title"
							}, {
								$bind: "description"
							}, {
								$bind: "gadgetType"
							}, {
								$bind: "icon"
							}, {
								$bind: "isFactory"
							}, {
								$bind: "applicationMenu"
							}]
						}
					}, {
						$category: "section",
						$title: "{@ContentSectionTitle}",
						$layout: {
							$items: [{
								$bind: "application"
							}, {
								$bind: "endpoint"
							}, {
								$bind: "representationRef"
							}, {
								$bind: "action"
							}, {
								$bind: "mobileDashboard"
							}, {
								$bind: "externalUrl"
							}, {
								$bind: "requestName"
							}, {
								$bind: "requestLevel"
							}, {
								$bind: "statName"
							}, {
								$bind: "processName"
							}, {
								$bind: "processMenu"
							}, {
								$bind: "processLeg"
							}]
						}
					}]
				}, {
					$items: [{
						$category: "section",
						$title: "{@ParametersSectionTitle}",
						$layout: {
							$items: [{
								$bind: "keyParameter"
							}, {
								$bind: "parameters"
							}]
						}
					}]
				}]

			}
		},
		$edit: {
			$copy: "$details"
		},
		$query: {
			$layout: {
				$items: [{
					$category: "section",
					$layout: {
						$items: [{
							$category: "section",
							$layout: {
								$items: [{
									$bind: "$resources",
									$layout: {
										$items: [{
											$bind: "code",
										}, {
											$bind: "title",
										}, {
											$bind: "description"
										}, {
											$bind: "gadgetType"
										}, {
											$bind: "entity"
										}, {
											$bind: "representation"
										}, {
											$bind: "fusionFunction"
										}, {
											$bind: "requestName"
										}, {
											$bind: "requestLevel"
										}, {
											$bind: "statName"
										}, {
											$bind: "processName"
										}, {
											$bind: "processMenu"
										}, {
											$bind: "processLeg"
										}, {
											$bind: "dashboard"
										}, {
											$bind: "application"
										}, {
											$bind: "endpoint"
										}, {
											$bind: "isFactory"
										}]
									}
								}]
							}
						}]
					}
				}]
			}
		},
		$lookup: {
			$copy: "$query"
		},
		$select: {
			$copy: "$query"
		}
	}
};
