// syracuse-ui/lib/tests/field/editor.js
"use strict";
var _fieldTester = require("syracuse-ui/lib/tests/field/fieldTester");
exports.load = function(page) {
	var $props = page.$prototype.$properties;
	$props.$field$rows = {
		$title: "$rows",
		$type: "application/x-string",
	};
	$props.$field$output = {
		$title: "output",
		$type: "application/x-string",
	};
	$props.textrtf = {
		$title: "text-rtf",
		$type: "text/rtf",
		$maxLength: 500,
		$capabilities: "sort,filter,raw",
		$links: {
			$tunnel: {},
			$lookup: {}
		}
	};
	$props.texthtml = {
		$title: "text-html",
		$type: "text/html",
		$links: {
			$tunnel: {},
			$lookup: {}
		}
	};
	$props.textplain = {
		$title: "text-plain",
		$type: "text/plain",
		$links: {
			$tunnel: {},
			$lookup: {}
		}
	};
	page.setTitle("Editor Field");
	_fieldTester.load(page);
	page.addFieldBlock(page.leftCol, "textplain", ["textplain"]);
	var box = page.addBox("textrtf", "h2");
	page.leftCol.appendChild(box.slot);
	page.loadNewItem(page.addCol(box.body, "50%"), {
		$bind: "textrtf",
		$rows: 5,
		$isEditMode: true
	});
	var right = page.addCol(box.body, "50%");
	page.loadNewItem(right, {
		$bind: "textrtf"
	});
	page.addFieldBlock(page.leftCol, "texthtml", ["texthtml"], function($edit) {
		$edit.$rows = 5;
	});

	page.addMetaDataBox();
	page.addMetaSectionField("$field$rows");
	page.addMetaSectionField("$field$output");
	page.applyChange({
		textplain: "Hello World!!!",
		textrtf: "{\\rtf1\\ansi{\\fonttbl{\\f0 Bitstream Charter;}} {\\f0 \\fs24 Ceci est un texte accentu\\u233 }\\par \\par {\\f0 des \\fs18 petits}{\\f0  et des \\fs32 gros}\\f0 .}",
		texthtml: "<h1>Hi</h1><p>This is very minimal \"hello world\" HTML document.</p>"
	});
};

exports.notifyDataChange = function(field, value) {
	var rtfField = field.page.boundFields["textrtf"][0],
		msg = "textrtf: caret=" + rtfField.getCaretPosition() + ", data=" + rtfField.getInputValue();
	field.page.boundFields["$field$output"][0].setDataValue(msg);
};
