// syracuse-ui/lib/tests/field/fieldTester.js
"use strict";

function _applyFieldMetaData(page, $bind, value, $field) {
	$field = $field || page.$prototype.$properties[$bind];
	$bind = $bind.replace("$field", "");
	var metaData = {};
	metaData[$bind] = value;
	if ($field.$isMetaData) {
		var $properties = {};
		Object.keys(page.boundFields).forEach(function(bind) {
			if (bind.indexOf("$field") != 0) {
				$properties[bind] = metaData;
			}
		});
		page.applyChange({
			$properties: $properties
		});
	} else {
		Object.keys(page.boundFields).forEach(function(bind) {
			if (bind.indexOf("$field") != 0) {
				page.boundFields[bind].forEach(function(field) {
					field.applyDesignMetaData(metaData, true);
				});
			}
		});
	}
}

exports.load = function(page) {
	page.notifyDiagnoses = function(value, useMetaDataOnly) {
		var self = this;
		var dataRecordSet = {
			$properties: {}
		};
		Object.keys(self.boundFields).forEach(function(bind) {
			var metaData = self.fillDiagnoses(value, bind);
			if (bind.indexOf("$field") != 0) {
				var $fieldProperty = self.$prototype.$properties[bind];
				dataRecordSet.$properties[bind] = {
					$diagnoses: metaData.$diagnoses
				};
			}
		});
		self.applyChange(dataRecordSet);
	};
	page.notifyActionLinkChange = function(metaData, useMetaDataOnly) {
		var dataRecordSet = {
			$properties: {}
		};
		Object.keys(this.boundFields).forEach(function(bind) {
			if (bind.indexOf("$field") != 0) {
				dataRecordSet.$properties[bind] = {
					$links: metaData.$links,
					$actions: metaData.$actions
				};
			}
		});
		this.applyChange(dataRecordSet);
	};
	page.fillDiagnoses = function(value, $bind) {
		var metaData = {};
		if (value == "clear") {
			metaData.$diagnoses = null;
		} else {
			metaData.$diagnoses = [{
				severity: "error",
				message: "Error test  blalfoz"
			}];
			if (value == "full") {
				metaData.$diagnoses.push({
					severity: "warning",
					message: "Warning test"
				});
				metaData.$diagnoses.push({
					severity: "info",
					message: "Info test"
				});
			}
		}
		return metaData;
	};
	page.notifyChangeToServer = function(sendBag) {
		var dataRecords = sendBag.dataRecords;
		delete sendBag.dataRecords;
		sendBag.dataRecords = dataRecords;
	};
	page.addMetaDataBox = function() {
		this.metaDataBox = this.addBox("MetaData", "h2");
		this.rightCol.appendChild(this.metaDataBox.slot);

		var $props = this.$prototype.$properties;
		$props.$field$isDiagnoses = {
			$type: "application/x-choice",
			$value: {
				$type: "application/x-string",

				$enum: [{
					$value: "full",
					$title: "error, warning, info"
				}, {
					$value: "error",
					$title: "default error"
				}, {
					$value: "clear",
					$title: "clear"
				}, ]
			}
		};
		$props.$field$description = {
			$type: "application/x-boolean",
			$title: "$description",
			$isMetaData: true
		};
		$props.$field$help = {
			$type: "application/x-boolean",
			$title: "$help",
			$isMetaData: true
		};
		$props.$field$isReadOnly = {
			$type: "application/x-boolean",
			$title: "$isReadOnly",
			$isMetaData: true
		};
		$props.$field$isEditMode = {
			$type: "application/x-boolean",
			$title: "$isEditMode",
			$isMetaData: true
		};
		$props.$field$isHidden = {
			$type: "application/x-boolean",
			$title: "$isHidden",
			$isMetaData: true
		};
		$props.$field$isDisabled = {
			$type: "application/x-boolean",
			$title: "$isDisabled",
			$isMetaData: true
		};
		$props.$field$isMandatory = {
			$type: "application/x-boolean",
			$title: "$isMandatory",
			$isMetaData: true
		};
		$props.$field$isTopLabelAlignment = {
			$type: "application/x-boolean",
			$title: "$isTopLabelAlignment"
		};
		$props.$field$isTitleHidden = {
			$type: "application/x-boolean",
			$title: "$isTitleHidden"
		};
		$props.$field$isRightTextLabelAlignment = {
			$type: "application/x-boolean",
			$title: "$isRightTextLabelAlignment"
		};
		$props.$field$setNull = {
			$type: "application/x-boolean",
			$title: "set Null"
		};
		$props.$field$LinksTypeMetaDataOnly = {
			$type: "application/x-boolean",
			$title: "use Meta Data Only",
			$isReadOnly: true,
			$isMetaData: true
		};
		$props.$field$LinksType = {
			$title: "$links $actions",
			$isMetaData: true,
			$type: "application/x-choice",
			$value: {
				$type: "application/x-string",

				$enum: [{
					$value: "$empty",
					$title: "$links = null"
				}, {
					$value: "$lazyload",
					$title: "$lazyload (transition)"
				}, {
					$value: "$links",
					$title: "$links"
				}, {
					$value: "$actions",
					$title: "$actions"
				}, {
					$value: "$linksParameters",
					$title: "$links with Parameters"
				}, {
					$value: "$actionsParameters",
					$title: "$actions with Parameters"
				}]
			}
		};
		this.$prototype.$field$LinksTypeMetaDataOnly = true;
		var metadataSection = ["$field$isHidden", "$field$isDisabled", "$field$isMandatory", "$field$isEditMode", "$field$isReadOnly", "$field$description", "$field$help", "$field$isTopLabelAlignment", "$field$isTitleHidden", "$field$isRightTextLabelAlignment", "$field$setNull"];
		metadataSection = metadataSection.concat(["$field$LinksTypeMetaDataOnly", "$field$LinksType", "$field$isDiagnoses"]);
		for (var ii = 0, jj = metadataSection.length; ii < jj; ii++) {
			this.addMetaSectionField(metadataSection[ii]);
		}
	};
	page.notifyDataChangeHandlers.push(function(field, value) {
		var page = field.page;
		switch (field.$item.$bind) {
			case "$field$setNull":
				Object.keys(page.boundFields).forEach(function(bind) {
					if (bind.indexOf("$field") != 0) {
						page.boundFields[bind].forEach(function(field) {
							field.setDataBind(null);
						});
					}
				});
				break;
			case "$field$LinksType":
				var metaData = page.getResources().getActionLinks(value);
				if (value == "$lazyload") {
					page.notifyDataChange(field, "$empty");
				}
				page.notifyActionLinkChange(metaData, page.boundFields.$field$LinksTypeMetaDataOnly[0].getDataValue());
				break;
			case "$field$help":
				_applyFieldMetaData(page, field.$item.$bind, value ? "description msdmdlsjgq" : null);
				break;
			case "$field$description":
				_applyFieldMetaData(page, field.$item.$bind, value ? "description df fmsdmdlsjgq" : null);
				break;
			case "$field$isDiagnoses":
				page.notifyDiagnoses(value, page.boundFields.$field$LinksTypeMetaDataOnly[0].getDataValue());
				break;

			default:
				if (field.$item.$bind.indexOf("$field") == 0) {
					switch (field.$item.$bind) {
						case "":
							break;
						default:
							_applyFieldMetaData(page, field.$item.$bind, value);
							break;
					}
				} else {
					return true; //do default notify	
				}
		}
	});
};
