// syracuse-ui/lib/tests/field/simple.js
"use strict";
var _fieldTester = require("syracuse-ui/lib/tests/field/fieldTester");

exports.load = function(page) {
	var $props = page.$prototype.$properties;
	$props.unknow = {
		$title: "unknow",
		$type: "application/x-unknow",

		$maxLength: 30,
		$capabilities: "sort,filter",
		$links: {
			$tunnel: {},
			$lookup: {}
		}
	};
	$props.xstring = {
		$title: "xstring",
		$type: "application/x-string",
		$maxLength: 30,
		$capabilities: "sort,filter",
		$links: {
			$tunnel: {},
			$lookup: {}
		}
	};
	$props.xstring$localizable = {
		$title: "Localizable property",
		$type: "application/x-string",
		$capabilities: "localize",
		$links: {
			$localize: {
				$title: "Translations",
				$url: "/syracuse-ui/lib/tests/field/localizablePropertyTest.json"
			}
		}
	};
	$props.xstring$localizablePicker = {
		$title: "Localizable property (picker)",
		$type: "application/x-string",
		$capabilities: "localize",
		$links: {
			$localize: {
				$title: "Translations",
				$url: "/syracuse-ui/lib/tests/field/localizablePropertyTest.json"
			}
		}
	};
	$props.xchoice = {
		$title: "xchoice",
		$type: "application/x-choice",
		$value: {
			$type: "application/x-integer",
			$enum: [0, 1, 2, 3, 4].map(function(value) {
				return {
					$value: value,
					$title: "Value " + value
				};
			})
		},
		$links: {
			$details: {
				$title: "couc"
			},
			$tunnel: {},
			$lookup: {}
		}
	};
	$props.xdate = {
		$title: "xdate",
		$type: "application/x-date",
		$links: {
			$tunnel: {},
			$lookup: {}
		}
	};
	$props.xtime = {
		$title: "xtime",
		$type: "application/x-time"
	};
	$props.xdatetime = {
		$title: "xdatetime",
		$type: "application/x-datetime"
	};
	$props.xboolean = {
		$title: "xboolean",
		$type: "application/x-boolean"
	};
	$props.$email = {
		$title: "$format: $email",
		$type: "application/x-string",
		$format: "$email"

	};
	$props.$phone = {
		$title: "$format: $phone",
		$type: "application/x-string",
		$format: "$phone"
	};
	$props.xinteger = {
		$title: "xinteger",
		$type: "application/x-integer",
		$isNullable: true,
		"$capabilities": "sort,filter",
		$links: {
			$tunnel: {}
		}
	};
	$props.xProgress = {
		$title: "xProgress",
		$type: "application/x-integer",
		$mode: "progress"
	};
	$props.xdecimal = {
		$title: "xdecimal",
		$type: "application/x-decimal",
		"$isNullable": true
	};
	$props.xquantity1 = {
		$title: "xquantity with xreference",
		$type: "application/x-quantity",
		$value: {
			$title: "{$value} {$unit}",
			$type: "application/x-decimal",
			$format: "0.00"
		},
		$unit: "CURRENCY"
	};
	$props.xquantity2 = {
		$title: "xquantity with xchoice",
		$type: "application/x-quantity",
		$value: {
			$title: "{$value} {$unit}",
			$type: "application/x-decimal",
			$format: "0.00"
		},
		$unit: "STATUS"
	};
	$props.YCUR = {
		"$type": "application/x-string",
		"$maxLength": 3,
		"$isHidden": true
	};
	$props.CURRENCY = {
		"$title": "{@323}",
		"$shortTitle": "{@323}",
		"$type": "application/x-reference",
		"$isMandatory": true,
		"$capabilities": "sort,filter",
		"$item": {
			"$url": "{$baseUrl}/TABCUR('{$key}')?representation=TABCUR.$lookup",
			"$value": "{CUR}",
			"$key": "{CUR}",
			"$properties": {
				"CUR": {
					"$type": "application/x-string"
				}
			},
			"CUR": "{YCUR}",
			"$links": {
				"$details": {
					"$title": "{@28123}",
					"$target": "",
					"$type": "application/json;vnd.sage=syracuse",
					"$url": "{$baseUrl}/TABCUR('{YCUR}')?representation=TABCUR.$details"
				},
				"$lookup": {
					"$title": "{@6109}",
					"$type": "application/json;vnd.sage=syracuse",
					"$url": "{$baseUrl}/TABCUR?representation=TABCUR.$lookup"
				},
				"$summary": {
					"$title": "{@25036}",
					"$target": "",
					"$type": "application/json;vnd.sage=syracuse",
					"$url": "{$baseUrl}/TABCUR('{YCUR}')?representation=TABCUR.$summary"
				},
				"$query": {
					"$title": "{@27982}",
					"$target": "",
					"$type": "application/json;vnd.sage=syracuse",
					"$url": "{$baseUrl}/TABCUR?representation=TABCUR.$query"
				}
			}
		}
	};
	$props.STATUS = {
		"$title": "{@23269}",
		"$shortTitle": "{@23269}",
		"$type": "application/x-choice",
		"$capabilities": "sort,filter",
		"$value": {
			"$type": "application/x-integer",
			"$enum": [{
				"$value": 1,
				"$title": "{@mes617_1}"
			}, {
				"$value": 2,
				"$title": "{@mes617_2}"
			}]
		}
	};
	$props.xreference = {
		$pluralType: "employees",
		$representation: "employee",
		$title: "xreference",
		$isMandatory: false,
		$capabilities: "sort,filter",
		$type: "application/x-reference",
		$url: "{$baseUrl}/{$pluralType}('{$key}')?representation={$representation}.$thumb",
		"$item": {
			$value: "{code}",
			$key: "{$uuid}",
			"$properties": {
				"code": {
					"$type": "application/x-string"
				},
				"description": {
					"$type": "application/x-string"
				}
			},
			$links: {
				"$details": {
					"$type": "application/json;vnd.sage=syracuse",
					"$url": "/sdata/syracuse/collaboration/syracuse/users('{$uuid}')?representation=country.$details&role={$role}"
				},
				"$lookup": {
					"$type": "application/json;vnd.sage=syracuse",
					"$url": "/sdata/syracuse/collaboration/syracuse/users?representation=country.$lookup&role={$role}&trackingId={$trackingId}&binding=country"
				},
				"$details1": {
					"$type": "application/json;vnd.sage=syracuse",
					"$title": "test link1",
					"$url": "/sdata/syracuse/collaboration/syracuse/users('{$uuid}')?representation=country.$details&role={$role}"
				},
				"$details2": {
					"$type": "application/json;vnd.sage=syracuse",
					"$title": "test link2",
					"$url": "/sdata/syracuse/collaboration/syracuse/users('{$uuid}')?representation=country.$details&role={$role}"
				},
			}
		}
	};
	$props.image = {
		$title: "image",
		$type: "image",
		$isReadOnly: true,
		$isDisabled: true
	};
	page.setTitle("Fields");
	_fieldTester.load(page);
	page.addFieldBlock(page.leftCol, "xstring", ["unknow", "notdefined", "xstring", "$phone", "$email"]);
	page.addFieldBlock(page.leftCol, "xboolean", ["xdate", "xtime", "xdatetime"]);
	page.addFieldBlock(page.leftCol, "Numeric", ["xinteger"]);
	page.addFieldBlock(page.leftCol, "Localizable", ["xstring$localizable", "xstring$localizablePicker"], function($edit) {
		if ($edit.$bind == "xstring$localizablePicker") {
			$edit.$useLocalizePicker = true;
		}
	});
	page.addFieldBlock(page.leftCol, "Progress", ["xProgress"]);
	page.addFieldBlock(page.leftCol, "Decimal", ["xdecimal"]);
	page.addFieldBlock(page.leftCol, "Quantity", ["xquantity1", "xquantity2"]);
	page.addFieldBlock(page.leftCol, "xboolean", ["xboolean"]);
	page.addFieldBlock(page.leftCol, "xboolean", ["xchoice", "xchoice", "xchoice"]);
	var xchoices = page.boundFields.xchoice;
	for (var ii = 0; ii < 6; ii++) {
		if (ii < 2) {
			xchoices[ii].setTitle("default >4 combo");
		} else
		if (ii < 4) {
			xchoices[ii].setTitle("$format:$combo");
			xchoices[ii].applyDesignMetaData({
				$format: "$combo"
			});
		} else {
			xchoices[ii].setTitle("$format:$radios");
			xchoices[ii].applyDesignMetaData({
				$format: "$radios"
			});
		}
	}
	page.addFieldBlock(page.leftCol, "xreference", ["xreference"]);
	page.addFieldBlock(page.leftCol, "image", ["image"]);

	page.addMetaDataBox();

	var ids = Object.keys(page.idMap);
	for (var ii = 0, jj = ids.length; ii < jj; ii++) {
		var item = page.idMap[ids[ii]];
		if (item && item.$item && item.$item.$bind == "notdefined") {
			item.isItemHidden = false;
			item.showItem(true);
		}

	}

	page.applyChange({
		"xstring": "ceci est du texte",
		"xstring$localizable": "Bague argent",
		"xstring$localizablePicker": "Bague argent",
		"$phone": "01.41.66.21.21",
		"$email": "toto@sage.com",
		"xinteger": 200,
		"xdecimal": 2987111111111.4548,
		"xboolean": true,
		"xdate": "2011-06-07",
		"xtime": "12:00:00",
		"xdatetime": "2012-03-16T12:40:50.957Z",
		"xchoice": 2,
		"xreference": {
			"$uuid": "4e54a41e-5c7d-4926-ad57-6a5f882c24c4",
			"$key": "4e54a41e-5c7d-4926-ad57-6a5f882c24c4",
			"$etag": 1,
			"$properties": {},
			"code": "AD",
			"description": "Andorradfd fsdqfsqdfsdq sdqf sdqfsdqf sdqqfsdq fsdqf sdqf sdfsdqf sdf sqdfsd qfsdq fsdqf sdqfsd qfsdq fsqd fs q fqsf",
			"$description": "Andorradfd fsdqfsqdfsdq sdqf sdqfsdqf sdqqfsdq fsdqf sdqf sdfsdqf sdf sqdfsd qfsdq fsdqf sdqfsd qfsdq fsqd fs q fqsf"
		},
		"YCUR": "USD",
		"CURRENCY": {
			"$symbol": "$"
		},
		"STATUS": 1,
		"xquantity1": 78.5,
		"xquantity2": 78.5
	});
};
