// @sage/syracuse-react/src/section/sectionHeader.tsx
"use strict";
import * as React from 'react';
import { Collapsable, Maximizable, Closeable } from './sectionLinks'

export interface ISectionHeaderProps {
    className: string,
    title: string,
    onCollapse?: () => void,
    collapsed?: boolean,
    onMaximize?: () => void,
    maximized?: boolean,
    onClose?: () => void
}

export class SectionHeader extends React.Component<ISectionHeaderProps, {}> {
    render() {
        let cssLink = this.props.className + "_btn";
        return <div className={this.props.className + "_header"}>
            <div className={this.props.className + "_title"}>{this.props.title}</div>
            {this.props.onCollapse && <Collapsable className={cssLink} collapsed={this.props.collapsed} onClick={this.props.onCollapse} />}
            {this.props.onMaximize && <Maximizable className={cssLink} maximized={this.props.maximized} onClick={this.props.onMaximize} />}
            {this.props.onClose && <Closeable className={cssLink} onClick={this.props.onClose} />}
        </div>;
    }
}
