// @sage/syracuse-react/src/section/sectionLinks.tsx
"use strict";
import * as React from 'react';
import { Link, LinkClickHandler } from '../link/link';
import  {getLocalString} from '../culture/localString';

interface ISectionLinkProps {
	className: string,
	onClick: LinkClickHandler
}


interface IMaximizableProps extends ISectionLinkProps {
	maximized: boolean
}
export class Maximizable extends React.Component<IMaximizableProps, {}>{
	render() {
		return <Link
			className={this.props.className}
			icon={this.props.maximized ? "minimize" : "maximize"}
			title={this.props.maximized ? getLocalString("section_minimize") : getLocalString("section_maximize")}
			onClick={this.props.onClick} />;
	}
}

export class Closeable extends React.Component<ISectionLinkProps, {}>{
	render() {
		return <Link
			className={this.props.className}
			icon="close"
			title={getLocalString("section_close")}
			onClick={this.props.onClick} />;
	}
}

interface ICollapsableProps extends ISectionLinkProps {
	collapsed: boolean
}
export class Collapsable extends React.Component<ICollapsableProps, {}>{
	render() {
		return <Link
			className={this.props.className}
			icon={this.props.collapsed ? "dir_down" : "dir_up"}
			title={this.props.collapsed ? getLocalString("section_expand") : getLocalString("section_collapse")}
			onClick={this.props.onClick} />;
	}
}
