// syracuse-ui/lib/field/chart/authoring/tabularChartAuthoring.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Article = require("syracuse-ui/lib/article/article").Article;


var _localize = {
    itemsPerPage: "page size"
};

function TabularChartAuthoring(){

}

exports.TabularChartAuthoring = helpers.defineClass(TabularChartAuthoring, Article, {
    notifyChange: function($bind, value){
        var metaData = {};
        metaData[$bind] = value;
        this.designedField.applyDesignMetaData(metaData, true);
        if (document.site.authoringView) {
            document.site.authoringView.notifyUpdate();
        }
    },
    applyChange: function(newData){
        if (newData) {
            // code to update authoring article
        }
        Article.prototype.applyChange.call(this, newData);
        
    },
    loadBox: function(initData){
        this.$prototype = {
            "$properties": {
                $chartType: this.designedField._createChoices()
            },
            $chartType: "column"
        };
        this.$item = {
            $category: "section",
            $layout: {
                $items: [{
                    $bind: "$chartType",
                    $icon: {
                        $mode: "iconText",
                        $path: "graphics/s-graph-",
                        $width: "32px",
                        $height: "32px"
                    },
                    $choiceLayout: "3",
                    $isEditMode: true,
                    $skin: "s-aw-field",
                    $isLeftSpaceHidden: true
                }]
            }
        };
        Article.prototype.loadBox.call(this, initData);
    },
    dispose: function(){
        Article.prototype.dispose.call(this);
    }
});
