// syracuse-ui/lib/field/chart/authoring/tabularChartAuthoring.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Article = require("syracuse-ui/lib/article/article").Article;


var _localize = {
    itemsPerPage: "page size"
};

function TabularChartAuthoring(){

}

exports.TabularChartAuthoring = helpers.defineClass(TabularChartAuthoring, Article, {
    onNotifyAuthoringChange: function(field, value, node){
        var metaData = {};
        metaData[field.$item.$bind] = value;
        this.designedField.applyDesignMetaData(metaData, true);
    },
    applyChange: function(newData){
        if (newData) {
            // code to update authoring article
        }
        Article.prototype.applyChange.call(this, newData);
        
    },
    loadBox: function(initData){
        this.$prototype = {
            "$properties": {
                $style: this.designedField._createChoices()
            },
            $style: "column"
        };
        this.$item = {
            $category: "section",
            $layout: {
                $items: [{
                    $bind: "$style",
                    $icon: {
                        $mode: "iconText",
                        $path: "graphics/s-graph-"
                    },
                    $choiceLayout: "3",
                    $isEditMode: true,
                    $skin: "s-aw-field",
                    $isLeftSpaceHidden: true
                }]
            }
        };
        Article.prototype.loadBox.call(this, initData);
    },
    dispose: function(){
        Article.prototype.dispose.call(this);
    }
});
