// @sage/syracuse-react/src/common/splitter/splitterOpener.tsx
"use strict";
import * as React from 'react';
import * as ReactDom from 'react-dom';
import { Link, LinkClickHandler } from '../../link/link';
import  {getLocalString} from '../../culture/localString';

export type Position = "left" | "right";
export const SlotId = "s_app_page_left_bar_btn_slot";

export class SplitterOpener extends React.Component<{
	position?: Position,
	collapsed: boolean,
	floating: boolean,
	onClick: LinkClickHandler
}, {}> {
	public root: HTMLElement;

	componentDidMount() {
		this.root = document.getElementById(SlotId);
		this.root.style.display = "inline-block";
		this.renderButton();
	}
	componentDidUpdate() {
		this.renderButton();
	}
	componentWillUnmount() {
		if (this.root) {
			ReactDom.unmountComponentAtNode(this.root);
			this.root.style.display = "none";
			this.root = null;
		}
	}
	renderButton() {
		let collapsed = this.props.collapsed;
		ReactDom.render(<Link className="s_profile_bar_iconlink"
			title={collapsed ? getLocalString("site_page_nav_show") : getLocalString("site_page_nav_hide")}
			icon={collapsed ? "menu" : "close"}
			onClick={this.props.onClick} />, this.root);
	}
	render() {
		return <noscript />;
	}
}
