// syracuse-ui/lib/desktop/field/collection/builder/arrayRecord.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Article = require("syracuse-ui/lib/common/article/article").Article;

function ArrayRecord(){
}

exports.ArrayRecord = helpers.defineClass(ArrayRecord, Article, {
    loadRecord: function(list, record, $recordIndex, isInsert){
        this.$recordUuid = record.$uuid;
        this.$recordIndex = $recordIndex;
        this.$facet = list.$recordFacet;
        this.$isEditMode = list.$isEditMode;
        this.$prototype = {
            $: {
                $bindRecord: list.$prototype.$item
            }
        };
        document.itemFactory.initializeItem(this, {}, list);
        this.isInsert = isInsert;
        this.loadBox(record);
    },
    getListDataSet: function(){
        return this.boxParent.getArticleParent().dataset[this.boxParent.$item.$bind];
    },
    onNotifyRecordChange: function(value, binding){
        var dataset = this.getListDataSet();
        dataset[this.$recordIndex] = value;
        this._notifyChange(dataset);
        return null;
    },
    _notifyChange: function(newDataSet){
        document.controller.notifyChange(this.boxParent.getArticleParent(), this.boxParent.$item.$bind, newDataSet);
    },
    remove: function(){
        this.$$item.remove();
    },
    _appendCapabilities: function(){
        var self = this;
        var list = self.boxParent;
        ["insert", "delete"].forEach(function($capability){
            if (list.builder.$capabilities.indexOf($capability) >= 0) {
                var btn = document.createElement("a");
                btn.className = list.$skin + "-record-" + $capability;
                btn.setAttribute("data-s-action", $capability);
                if (!self.$$actions) {
                    self.$$actions = $(document.createElement("div")).addClass(list.$skin + "-record-actions").prependTo(self.$$value);
                }
                $(btn).appendTo(self.$$actions);
            }
        });
        self.$$item.delegate("a[data-s-action]", "click", function(){
            list.builder.doCapabilities(self, this.getAttribute("data-s-action"));
            return false;
        });
    },
    drawBox: function(){
        var list = this.boxParent;
        var div = document.createElement("div");
        div.className = list.$skin + "-record-item";
        div.setAttribute("data-s-record", this.$recordIndex);
        this.$$item = $(div);
        
        div = document.createElement("div");
        div.className = list.$skin + "-record-value";
        this.$$value = $(div).appendTo(this.$$item);
        
        div = document.createElement("div");
        div.className = list.$skin + "-record-data";
        var $$data = $(div).appendTo(this.$$value);
        if (this.$isEditMode) {
            if (list.builder.$capabilities) {
                this._appendCapabilities();
            }
        }
        document.itemFactory.load($$data, {
            $isTitleRowHidden: true,
            $bind: "$bindRecord",
            $isEditMode: this.$isEditMode,
            $skin: "s-array-item",
            $inplace: true
        }, this);
        if (this.isInsert) {
            this.$$item.insertBefore(list.dataRows[this.$recordIndex].$$item);
            delete this.isInsert;
        }
        else {
            this.$$item.appendTo(list._$$body);
        }
    },
    dispose: function(){
        if (this.$$item) {
            this.$$item.undelegate();
        }
        delete this.menus;
        Article.prototype.dispose.call(this);
    }
});
