// syracuse-ui/lib/desktop/field/collection/builder/arrayBuilder.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var RawListBuilder = require("syracuse-ui/lib/common/field/collection/rawListBuilder").RawListBuilder;
var ArrayRecord = require('./arrayRecord').ArrayRecord;

function ArrayBuilder(){
}

exports.Builder = helpers.defineClass(ArrayBuilder, RawListBuilder, {
    initialize: function(){
        this.list.$item.$isQuickDesignerEnabled = false;
        this.list.$item.$isPagerHidden = true;
    },
    appendRecord: function(record, $recordIndex, isInsert){
        var record = {
            $uuid: $recordIndex,
            $bindRecord: record
        };
        var recordArticle = (this.list.dataset[record.$uuid] = new ArrayRecord());
        recordArticle.$layoutOptions = this.$layoutOptions;
        recordArticle.loadRecord(this.list, record, $recordIndex);
        return recordArticle;
    },
    drawBuilder: function(){
        this.list.$skin = this.list.$item.$skin || "s-array";
        
        var div = document.createElement("div");
        div.className = this.list.$skin + "-body";
        this.list._$$body = $(div).appendTo(this.list.$$listSlot);
        
        this.list.applyDesignMetaData(this.list.$item, false);
        this.list.setState(this.list.$field);
        this.list.boxParent.getArticle().bind(this.list, this.list.$item.$bind);
    },
    setDataBind: function($resources, record, metaData){
        var redraw = false;
        var self = this;
        if (metaData) {
            redraw = self.list.applyMetaData(metaData);
        }
        if ($resources !== undefined) {
            self.list.applySettings(record);
            var body = self.list._$$body[0];
            body.style.display = "none";
            if (redraw) {
                self.onRedrawCore();
            }
            self.list.removeRecords();
            if ($resources) {
                $resources.forEach(function(record, $recordIndex){
                    self.list.dataRows.push(self.appendRecord(record, $recordIndex));
                });
            }
            if (self.list.$isEditMode) {
                debugger;
                if (self.list.$prototype.$minItems) {
                    while (self.list.dataRows.length < self.list.$prototype.$minItems) {
                        self.list.dataRows.push(self.appendRecord(null, self.list.dataRows.length));
                    }
                }
				if (self.list.$prototype.$maxItems) {
                    if(self.list.dataRows.length < self.list.$prototype.$maxItems) {
                        self.list.dataRows.push(self.appendRecord(null, self.list.dataRows.length));
                    }
                }
            }
            self.onAfterDataBinding();
            body.style.display = "";
        }
    },
    dispose: function(){
        RawListBuilder.prototype.dispose.call(this);
    }
});
