// syracuse-ui/lib/desktop/field/collection/builder/cardRecord.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var RecordArticle = require("./recordArticle").RecordArticle;

function CardRecord(){
}

exports.CardRecord = helpers.defineClass(CardRecord, RecordArticle, {
    initializeRecord: function(options){
        var rowIndex;
        if (this.builder.rows.length == 0) {
            this.row = this.builder.appendCardsRow();
        }
        else {
            this.row = this.builder.rows[this.builder.rows.length - 1];
        }
        if (this.list.$item.$cardsByRowCount) {
            if (this.row.count == this.list.$item.$cardsByRowCount) {
                this.row = this.builder.appendCardsRow();
            }
        }
        else {
            var div = document.createElement("div");
            div.className = this.list.$recordCss;
            div.style.visibility = "hidden";
            this.row.cells.push(row.slot.appendChild(div));
        }
        this.cellIndex = this.row.count++;
        options.$item = {
            $category: "section",
            $layout: {
                $items: this.list.$item.$layout.$items
            }
        };
    },
    remove: function(){
        this.$$item.remove();
    },
    drawBox: function(){
        var item = this.row.cells[this.cellIndex];
        item.setAttribute("data-s-record", this.$recordUuid);
        this.$$item = $(item);
        if (this.list.$item.$alternateStyle && this.$recordIndex % 2) {
            item.className += " " + this.list.$skin + "-alt";
        }
        var selectorSlot = item.appendChild(document.createElement("div"));
        
        var dom = document.createElement("div");
        dom.className = this.list.$skin + "-record";
        var $$data = $(item.appendChild(dom));
        if (!this.list.$item.$isMenuRecordHidden) {
            dom = document.createElement("div");
            dom.className = this.list.$skin + "-menu-row";
            this._loadMenu(item.appendChild(dom));
        }
        if (this.list.recordSelector.appendToRecord) {
            this._selectorCheck = this.list.recordSelector.appendToRecord(selectorSlot);
        }
        else {
            selectorSlot.style.display = "none";
        }
        document.itemFactory.load($$data, this.list.$item.$cardview, this);
        item.style.visibility = "";
    },
    highlightSelection: function(selected){
        document.site.toggleClass(this.$$item[0], "s-list-record-selected", selected);
        if (this._selectorCheck) {
            $(this._selectorCheck).attr("checked", selected);
        }
    },
    dispose: function(){
        delete this.row;
        RecordArticle.prototype.dispose.call(this);
    }
});
