// syracuse-ui/lib/desktop/field/collection/builder/cardListBuilder.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var CardRecord = require('./cardRecord').CardRecord;
var RawListBuilder = require("syracuse-ui/lib/common/field/collection/rawListBuilder").RawListBuilder;

function CardListBuilder(){
}

exports.Builder = helpers.defineClass(CardListBuilder, RawListBuilder, {
    initialize: function(){
        this.list.RecordClass = CardRecord;
    },
    applyDesignMetaData: function(metadata, onAuthoring){
        if (onAuthoring) {
            if (metadata.$cardsByRowCount !== undefined) {
                this.list.$item.$cardsByRowCount = metadata.$cardsByRowCount;
                this.list.reloadBuilder();
            }
            if (metadata.$graphMode !== undefined) {
                this.toggleDataList(metadata.$graphMode != "graph");
            }
        }
    },
    toggleDataList: function(show){
        if (this._$$body) {
            this._$$body[0].style.display = show ? "" : "none";
        }
        else {
            this._drawDataList();
        }
    },
    _drawDataList: function(){
        var list = this.list;
        list._$$body = $(document.createElement("div")).appendTo(list.$$listSlot);
        list.appendSelector();
        var css = list.$skin + ((list.$item.$cardsByRowCount == 0) ? "-free" : "-cell");
        list.$recordCss = css + "-record";
        list._$$body.addClass(css + "-body");
        if (!list.$item.$isMenuRecordHidden) {
            list.$$item.delegate("div[data-s-record]", "mouseenter mouseleave", function(event){
                var $uuid = this.getAttribute("data-s-record");
                list.toggleRecordMenu(event.type == "mouseenter", list.dataset[$uuid]);
                return false;
            });
        }
    },
    drawBuilder: function(){
        var list = this.list;
        this.rows = [];
        list.$skin = list.$item.$skin || "s-cards";
        
        var div = document.createElement("div");
        div.className = "s-list";
        list.$$list = $(div).appendTo(list.$$dataValue);
        
        div = document.createElement("div");
        div.className = "s-list-topbar";
        list.$$topbar = $(div).appendTo(list.$$list);
        list.appendArticleMenus(list.$$topbar);
        if (!list.$item.$hidePager) {
            list.appendPager(list.$$topbar);
        }
        list._appendCore();
        
        list._appendAuthoringOpener(list.$$topbar);
        
        list.$cardView = (typeof(list.$item.$cardView) === "object") ? list.$item.$cardView : {
            $layout: {
                $items: list.$item.$layout.$items
            }
        };
        if (list.$item.$cardsByRowCount == undefined) {
            list.$item.$cardsByRowCount = 1;
        }
        this._drawDataList();
        list.applyDesignMetaData(list.$item, false);
        list.boxParent.getArticle().bind(list, list.$item.$bind);
    },
    appendCardsRow: function(){
        var row = {
            index: this.rows.length,
            cards: {},
            count: 0,
            $$cells: [],
            $$row: $("<div/>").addClass(this.list.$recordCss + "-row").appendTo(this.list._$$body)
        };
        if (this.list.$item.$cardsByRowCount) {
            var count = this.list.$item.$cardsByRowCount || 1;
            var width = (100 / count) + "%";
            for (var ii = 0; ii < count; ii++) {
                row.$$cells.push($("<div/>").addClass(this.list.$recordCss).css("width", width).css("visibility", "hidden").appendTo(row.$$row));
            }
        }
        
        this.rows.push(row);
        return row;
    },
    appendRecord: function(record, $recordIndex){
        var row, rowIndex;
        if (this.rows.length == 0) {
            row = this.appendCardsRow();
        }
        else {
            row = this.rows[this.rows.length - 1];
        }
        if (this.list.$item.$cardsByRowCount) {
            if (row.count == this.list.$item.$cardsByRowCount) {
                row = this.appendCardsRow();
            }
        }
        else {
            var div = document.createElement("div");
            div.className = this.list.$recordCss;
            div.style.visibility = "hidden";
            row.$$cells.push($(div).appendTo(row.$$row));
        }
        var recordArticle = (this.list.dataset[record.$uuid] = new this.list.RecordClass());
        recordArticle.$layoutOptions = this.$layoutOptions;
        recordArticle.loadRecord(this.list, record, $recordIndex, row, row.count++);
        
    }
});
