// syracuse-ui/lib/desktop/field/collection/builder/cardListBuilder.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var CardRecord = require('./cardRecord').CardRecord;
var RawListBuilder = require("syracuse-ui/lib/common/field/collection/rawListBuilder").RawListBuilder;

function CardsBuilder(){
}

exports.Builder = helpers.defineClass(CardsBuilder, RawListBuilder, {
    initialize: function(){
        this.list.$item.$isTopLabelAlignment = true;
        this.list.$item.$cardsByRowCount = this.list.$item.$cardsByRowCount || 1;
        this.RecordClass = CardRecord;
        //this.list.recordSelector = "div[data-s-record]";
        RawListBuilder.prototype.initialize.call(this);
    },
    applyDesignMetaData: function(metadata, onAuthoring){
        if (onAuthoring) {
            if (metadata.$cardsByRowCount !== undefined) {
                this.list.$item.$cardsByRowCount = metadata.$cardsByRowCount;
                this.list.reloadBuilder();
            }
            if (metadata.$graphMode !== undefined) {
                this.toggleDataList(metadata.$graphMode != "graph");
            }
        }
    },
    toggleDataList: function(show){
        if (this._body) {
            this._body.style.display = show ? "" : "none";
        }
        else {
            this.list._core.appendChild(this._body = document.createElement("div"));
            this.list.loadRecordSelector();
            this.list.$recordCss = this.list.$skin + "-cell-record";
            this._body.className = this.list.$skin + "-cell-body";
        }
    },
    emptyListBody: function(addEmptySlot){
        RawListBuilder.prototype.emptyListBody.call(this);
        this.rows = [];
    },
    drawBuilder: function(){
        var list = this.list;
        list.$$fieldValue[0].style.overflow = "auto";
        this.rows = [];
        list.$skin = list.$item.$skin || "s-cards";
        
        list.$item.$cards = (typeof(list.$item.$cards) === "object") ? list.$item.$cards : {
            $layout: {
                $items: list.$item.$layout.$items
            }
        };
        list.applyDesignMetaData(list.$item, false);
        this.list.parseCapabilities(this.list.$prototype.$capabilities);
        this.toggleDataList(true);
        list.boxParent.getArticle().bind(list, list.$item.$bind);
    },
    appendCardsRow: function(){
        var row = {
            index: this.rows.length,
            count: 0,
            cells: [],
            max: this.list.$item.$cardsByRowCount || 1
        };
        row.slot = document.createElement("div");
        row.slot.className = this.list.$recordCss + "-row";
        this._body.appendChild(row.slot);
        var width = (100 / row.max) + "%";
        for (var ii = 0; ii < row.max; ii++) {
            var cell = document.createElement("div");
            cell.className = "s-cards-cell";
            cell.style.width = width;
            cell.visibility = "hidden";
            row.cells.push(row.slot.appendChild(cell));
        }
        this.rows.push(row);
        return row;
    }
});
