// syracuse-ui/lib/desktop/field/collection/builder/gridCardViewDecorator.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function GridCardViewDecorator(){

}

exports.GridCardViewDecorator = helpers.defineClass(GridCardViewDecorator, null, {
    initialize: function(builder){
        var self = this;
        self.list = (self.builder = builder).list;
        var $item = self.list.$item;
        $item.$cardview = (typeof($item.$cardview) === "object") ? $item.$cardview : {
            $layout: {
                $items: self.list.$binds.map(function($bind){
                    return {
                        $bind: $bind
                    };
                })
            }
        };
        $item.$cardview.$category = $item.$cardview.$category || "section"
        self.$cardviewPosition = $item.$cardviewPosition || "inline";
        if (self.$cardviewPosition != "inline") {
            if ($item.$selectMode && $item.$selectMode != "row") {
                self.$cardviewPosition = "inline";
            }
            else {
                $item.$selectMode = "row";
            }
        }
        if (self.$cardviewPosition != "inline") {
            var div = document.createElement("div");
            div.className = "s-list-gridcard-slot";
            var wrapper = (self.$$wrapper = $(div))[0];
            
            div = document.createElement("div");
            div.className = "s-list-gridcard-grid-slot";
            var tableSlot = (self.list.$$tableSlot = $(div).appendTo(self.$$wrapper))[0];
            
            var div = document.createElement("div");
            div.className = "s-list-gridcard-card-slot";
            var cardSlot = (self.$$cardSlot = $(div))[0];
            
            
            if (self.$cardviewPosition == "left" || self.$cardviewPosition == "right") {
                wrapper.style.display = "table";
                wrapper.style.tableLayout = "fixed";
                wrapper.style.width = "100%";
                tableSlot.style.display = "table-cell";
                cardSlot.style.display = "table-cell";
                /*var widths = (self.list.$item.$cardviewColumnWidth || "50,50").split(",");
                 cardSlot.style.width = widths[(self.$cardviewPosition == "left") ? 0 : 1] + "%";
                 tableSlot.style.width = widths[(self.$cardviewPosition == "left") ? 1 : 0] + "%";*/
                cardSlot.style.width = (self.list.$item.$cardviewWidth || "50") + "%";
            }
            else {
                wrapper.style.display = "";
                wrapper.style.tableLayout = "";
                wrapper.style.width = "";
                tableSlot.style.display = "";
                cardSlot.style.display = "";
                tableSlot.style.width = "";
                cardSlot.style.width = "";
            }
            self.$$cardSlot[(["left", "top"].indexOf(self.$cardviewPosition) >= 0) ? "prependTo" : "appendTo"](self.$$wrapper.appendTo(self.list.$$listSlot));
        }
    },
    release: function(){
        if (this.$$wrapper) {
            this.list.$$listSlot.append(this.list.$$tableSlot.children());
            this.list.$$tableSlot = this.list.$$listSlot;
            this.$$wrapper.remove();
            delete this.$$wrapper;
        }
    },
    onMakeColGroup: function(cols){
        if (this.$cardviewPosition == "inline") {
            cols += "<col/>";
        }
        return cols;
    },
    onAppendMenuRow: function(record, colSpan){
        if (this.$cardviewPosition == "inline") {
            colSpan--;
            record.appendEmptyMenuCell();
        }
        return colSpan;
    },
    onMakeTitleRow: function(){
        if (this.$cardviewPosition == "inline") {
            var th = document.createElement("th");
            th.className = this.builder.gridCss.title + " " + this.list.$skin + "-cardview-cell";
            this.builder.$$titleRow.append($(th).append("<a class='" + this.list.$skin + "-cardview-opener s-close'/>"));
        }
    },
    onDrawOrientation: function(){
        if (this.$cardviewPosition == "inline") {
            this.builder.$columnsCount++;
            this._bindOpener();
        }
    },
    onRecordDrawBox: function(recordArticle){
        if (this.$cardviewPosition == "inline") {
            var th = document.createElement("td");
            th.className = this.builder.gridCss.cell + " " + this.list.$skin + "-cardview-cell";
            var btn = document.createElement("a");
            btn.className = this.list.$skin + "-cardview-opener s-close";
            recordArticle.$$item.append($(th).append(recordArticle.$$cardviewOpener = $(btn)));
        }
    },
    _bindOpener: function(){
        var self = this;
        self.builder.list.$$item.delegate("a." + self.builder.list.$skin + "-cardview-opener", "click", function(){
            var $$opener = $(this);
            var show = $$opener.hasClass("s-close");
            var $uuid = $$opener.closest("[data-s-record]").attr("data-s-record");
            var $uuidList = ($uuid) ? [$uuid] : Object.keys(self.builder.list.dataset);
            $uuidList.forEach(function($uuid){
                self._toggleRecordCardview(self.builder.list.dataset[$uuid], show);
            });
            document.site.toggleClass($$opener, "s-close", !show);
        });
    },
    _toggleRecordCardview: function(recordArticle, show){
        recordArticle.$$cardviewOpener.toggleClass("s-close", !show);
        if (!recordArticle.$$cardviewSlot) {
            var td = document.createElement("td");
            td.className = recordArticle.boxParent.$skin + "-cardview-cell";
            td.setAttribute("colspan", recordArticle.boxParent.builder.$columnsCount);
            var $$cell = $(td).appendTo(recordArticle.$$cardviewSlot = $(document.createElement("tr")).insertAfter(recordArticle.$$item));
            this._createCardView($$cell, recordArticle);
        }
        recordArticle.$$cardviewSlot.toggle(show);
    },
    _createCardView: function($$container, recordArticle){
        this.cardview = document.itemFactory.load($$container, this.list.$item.$cardview, recordArticle);
    },
    onRecordHighlightSelection: function(recordArticle, selected){
        if (this.$cardviewPosition == "inline") {
            if (recordArticle.$$cardviewSlot) {
                document.site.toggleClass(recordArticle.$$cardviewSlot, "s-list-record-selected", selected);
            }
        }
        else {
            if (this.cardview) {
                recordArticle.removeItem(this.cardview, true);
                delete this.cardview;
            }
            if (selected) {
                this._createCardView(this.$$cardSlot, recordArticle);
            }
        }
    },
    onRecordRemove: function(recordArticle){
        if (recordArticle.$$cardviewSlot) {
            recordArticle.$$cardviewSlot.remove();
        }
    },
    onQuickFilterDrawBox: function(htmlCell, $$container){
        if (this.$cardviewPosition == "inline") {
            $(htmlCell).appendTo($$container);
        }
    },
    onAfterDataBinding: function(){
        if (this.$cardviewPosition != "inline" && Object.keys(this.list.selectedRecords).length == 0) {
            if (this.list.dataRows.length > 0) {
                this.list.selectRecord(this.list.dataRows[0].$recordUuid, true);
            }
        }
    },
    dispose: function(){
        delete this.list;
        delete this.cardview;
        delete this.builder;
    }
});
