// syracuse-ui/lib/desktop/field/collection/builder/gridCardViewDecorator.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function GridCardViewDecorator(){

}

exports.GridCardViewDecorator = helpers.defineClass(GridCardViewDecorator, null, {
    initialize: function(builder){
        var $item = (this.builder = builder).list.$item;
        this.$cardview = (typeof($item.$cardview) === "object") ? $item.$cardview : {
            $layout: {
                $items: builder.list.$binds.map(function($bind){
                    return {
                        $bind: $bind
                    };
                })
            }
        };
        this.$cardviewPosition = $item.$cardviewPosition || "inline";
        if (this.$cardviewPosition != "inline") {
            if ($item.$selectMode && $item.$selectMode != "row") {
                this.$cardviewPosition = "inline";
            }
            else {
                $item.$selectMode = "row";
            }
        }
        if (this.$cardviewPosition != "inline") {
            var div = document.createElement("div");
            div.className = "s-list-gridcard-slot";
            var wrapper = (this.$$wrapper = $(div))[0];
            
            div = document.createElement("div");
            div.className = "s-list-gridcard-grid-slot";
            var tableSlot = (builder.list.$$tableSlot = $(div).appendTo(this.$$wrapper))[0];
            
            var div = document.createElement("div");
            div.className = "s-list-gridcard-card-slot";
            var cardSlot = (this.$$cardSlot = $(div))[0];
            
            
            if (this.$cardviewPosition == "left" || this.$cardviewPosition == "right") {
                wrapper.style.display = "table";
                wrapper.style.tableLayout = "fixed";
                wrapper.style.width = "100%";
                tableSlot.style.display = "table-cell";
                cardSlot.style.display = "table-cell";
                var widths = (builder.list.$item.$cardviewColumnWidth || "50,50").split(",");
                cardSlot.style.width = widths[(this.$cardviewPosition == "left") ? 0 : 1] + "%";
                tableSlot.style.width = widths[(this.$cardviewPosition == "left") ? 1 : 0] + "%";
            }
            else {
                wrapper.style.display = "";
                wrapper.style.tableLayout = "";
                wrapper.style.width = "";
                tableSlot.style.display = "";
                cardSlot.style.display = "";
                tableSlot.style.width = "";
                cardSlot.style.width = "";
            }
            this.$$cardSlot[(["left", "top"].indexOf(this.$cardviewPosition) >= 0) ? "prependTo" : "appendTo"](this.$$wrapper.appendTo(builder.list.$$listSlot));
        }
    },
    release: function(){
        if (this.$$wrapper) {
            this.builder.list.$$listSlot.append(this.builder.list.$$tableSlot.children());
            this.builder.list.$$tableSlot = this.builder.list.$$listSlot;
            this.$$wrapper.remove();
            delete this.$$wrapper;
        }
    },
    makeColGroup: function(cols){
        if (this.$cardviewPosition == "inline") {
            cols += "<col/>";
        }
    },
    makeTitleRow: function(){
        if (this.$cardviewPosition == "inline") {
            var th = document.createElement("th");
            th.className = this.builder.gridCss.title + " " + this.builder.list.$skin + "-cardview-cell";
            this.builder.$$titleRow.append($(th).append("<a class='" + this.builder.list.$skin + "-cardview-opener s-close'/>"));
        }
    },
    drawOrientation: function(){
        if (this.$cardviewPosition == "inline") {
            this.builder.$columnsCount++;
            this._bindOpener();
        }
    },
    onRecordDrawBox: function(recordArticle){
        if (this.$cardviewPosition == "inline") {
            var th = document.createElement("td");
            th.className = this.builder.gridCss.cell + " " + this.builder.list.$skin + "-cardview-cell";
            var btn = document.createElement("a");
            btn.className = this.builder.list.$skin + "-cardview-opener s-close";
            recordArticle.$$item.append($(th).append(recordArticle.$$cardviewOpener = $(btn)));
        }
    },
    _bindOpener: function(){
        var self = this;
        self.builder.list.$$item.delegate("a." + self.builder.list.$skin + "-cardview-opener", "click", function(){
            var $$opener = $(this);
            var show = $$opener.hasClass("s-close");
            var $uuid = $$opener.closest("[data-s-record]").attr("data-s-record");
            var $uuidList = ($uuid) ? [$uuid] : Object.keys(self.builder.list.dataset);
            $uuidList.forEach(function($uuid){
                self._toggleRecordCardview(self.builder.list.dataset[$uuid], show);
            });
            document.site.toggleClass($$opener, "s-close", !show);
        });
    },
    _toggleRecordCardview: function(recordArticle, show){
        recordArticle.$$cardviewOpener.toggleClass("s-close", !show);
        if (!recordArticle.$$cardviewSlot) {
            var td = document.createElement("td");
            td.className = recordArticle.boxParent.$skin + "-cardview-cell";
            td.setAttribute("colspan", recordArticle.boxParent.builder.$columnsCount);
            var $$cell = $(td).appendTo(recordArticle.$$cardviewSlot = $(document.createElement("tr")).insertAfter(recordArticle.$$item));
            this._createCardView($$cell, recordArticle);
        }
        recordArticle.$$cardviewSlot.toggle(show);
    },
    _createCardView: function($$container, recordArticle){
        this.$cardview.$category = this.$cardview.$category || "section"
        this.cardview = document.itemFactory.load($$container, this.$cardview, recordArticle);
    },
    onRecordHighlightSelection: function(recordArticle, selected){
        if (this.$cardviewPosition == "inline") {
            if (recordArticle.$$cardviewSlot) {
                document.site.toggleClass(recordArticle.$$cardviewSlot, "s-list-record-selected", selected);
            }
        }
        else {
            if (this.cardview) {
                recordArticle.removeItem(this.cardview, true);
                delete this.cardview;
            }
            if (selected) {
                this._createCardView(this.$$cardSlot, recordArticle);
            }
        }
    },
    onRecordRemove: function(recordArticle){
        if (recordArticle.$$cardviewSlot) {
            recordArticle.$$cardviewSlot.remove();
        }
    },
    onQuickFilterDrawBox: function(htmlCell, $$container){
        if (this.$cardviewPosition == "inline") {
            $(htmlCell).appendTo($$container);
        }
    },
    onAfterDataBinding: function(){
        var list = this.builder.list;
        if (this.$cardviewPosition != "inline" && Object.keys(list.selectedRecords).length == 0) {
            if (list.dataRows.length > 0) {
                list.selectRecord(list.dataRows[0].$recordUuid, true);
            }
        }
    },
    dispose: function(){
        delete this.cardview;
        delete this.builder;
    }
});
