// syracuse-ui/lib/desktop/field/collection/builder/hgridBuilder.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var ListBuilder = require("./listBuilder").ListBuilder;
var GridCardViewDecorator = require("./gridCardViewDecorator").GridCardViewDecorator;
var Table = require("./table").Table;

var HGridRecord = require('./hgridRecord').HGridRecord;


function HGridBuilder(){
}

exports.Builder = helpers.defineClass(HGridBuilder, ListBuilder, {
    initialize: function(){
        this.$defaultSkin = "s-" + this.list.$item.$format;
        this.RecordClass = HGridRecord;
        this.list.$item.$isTopLabelAlignment = true;
        // this.list.recordSelector = "tr[data-s-record]";
        ListBuilder.prototype.initialize.call(this);
    },
    emptyListBody: function(addEmptySlot){
        var self = this;
        if (self.list.$item.$format == "vgrid") {
            self.table.orderCols.forEach(function(col){
                var children = $(col.slot).children();
                var dataColsCount = children.length - self.table.headerColsCount;
                if (dataColsCount) {
                    for (var ii = self.table.headerColsCount; ii < children.length; ii++) {
                        $(children[ii]).remove();
                    }
                }
            });
            if (addEmptySlot) {
                var emptySlot = document.createElement("td");
                emptySlot.className = self.list.$skin + "-empty-slot";
                emptySlot.setAttribute("rowspan", self.table.orderCols.length);
                self.table.orderCols[0].slot.appendChild(emptySlot);
            }
        }
        else {
            ListBuilder.prototype.emptyListBody.call(self);
            if (addEmptySlot) {
                var emptySlot = document.createElement("td");
                emptySlot.className = this.list.$skin + "-empty-slot";
                emptySlot.setAttribute("colspan", self.table.orderCols.length);
                self._body.appendChild(document.createElement("tr")).appendChild(emptySlot);
            }
        }
    },
    applyDesignMetaData: function(metaData, onAuthoring){
        if (onAuthoring) {
            if (metaData.$isRowIndexVisible !== undefined) {
                this.list.$item.$isRowIndexVisible = metaData.$isRowIndexVisible;
                this.list.reloadBuilder();
            }
            if (metaData.$graphMode !== undefined) {
                this.toggleDataList(metaData.$graphMode != "graph");
            }
            if (metaData.$cardsWidth !== undefined) {
                this.list.$item.$cardsWidth = metaData.$cardsWidth;
            }
            if (metaData.$cards !== undefined || metaData.$cardsPosition !== undefined) {
                if (metaData.$cards !== undefined) {
                    this.list.$item.$cards = metaData.$cards;
                }
                if (metaData.$cardsPosition !== undefined) {
                    this.list.$item.$cardsPosition = metaData.$cardsPosition;
                }
                if (this.cardview) {
                    this.cardview.release();
                }
                this.list.reloadBuilder();
            }
            if (metaData.$cardsWidth !== undefined) {
                if (this.cardview) {
                    this.cardview._cardSlot.style.width = this.list.$item.$cardsWidth + "%";
                }
            }
            if (metaData.$fitContainer !== undefined) {
                this.table._table.style.width = (this.list.$item.$fitContainer = metaData.$fitContainer) ? "100%" : "";
            }
        }
        
    },
    renderOrderBy: function(){
        this.table.renderOrderBy();
    },
    toggleDataList: function(show){
        if (this.table) {
            this.table.toggle(show);
        }
        else {
            (this.table = new Table()).load(this, this.list._tableSlot || this.list._core);
            this._body = this.table.appendBody();
            
            this.list.loadRecordSelector();
            this.onRedrawCore();
        }
    },
    ensureCardView: function(){
        var self = this
        var $collectionBinds = [];
        if (!self.list.$item.$cards) {
            self.list.parseItems(function($item, $field){
                if ($field && $field.$type == "application/x-array") {
                    if (!$field.$isHidden) {
                        if (self.list.$item.$cards) {
                            $field.$isHidden = true;
                        }
                        else {
                            $collectionBinds.push($item.$bind);
                        }
                    }
                }
            });
            if ($collectionBinds.length > 0) {
                self.list.$item.$layout.$items = self.list.$item.$layout.$items.filter(function($item){
                    return $collectionBinds.indexOf($item.$bind) < 0;
                });
                self.list.$item.$cards = {
                    $layout: {
                        $layoutType: "tabs",
                        $items: $collectionBinds.map(function($bind){
                            return {
                                $category: "section",
                                $title: self.list.$fields[$bind].$title,
                                $layout: {
                                    $items: [{
                                        $bind: $bind,
                                        $isTitleHidden: true
                                    }]
                                }
                            
                            };
                        })
                    }
                };
            }
        }
        if (self.list.$item.$cards) {
            self.cardview = new GridCardViewDecorator();
            self.cardview.initialize(self);
        }
    },
    drawBuilder: function(){
        var self = this;
        var list = this.list;
        list.$$fieldValue[0].style.overflow = "auto";
        var $skin = list.$skin = list.$item.$skin || this.$defaultSkin;
        this.gridCss = {
            title: $skin + "-title-cell",
            cell: $skin + "-cell",
            rowIndex: $skin + "-cell " + $skin + "-row-index"
        };
        
        self.ensureCardView();
        
        list.applyDesignMetaData(list.$item, false);
        list.setState(list.$field);
        this.list.parseCapabilities(this.list.$prototype.$capabilities);
        this.toggleDataList(true);
        list.boxParent.getArticle().bind(list, list.$item.$bind);
        this.table.toggle(true);
    },
    onAfterDataBinding: function(){
        ListBuilder.prototype.onAfterDataBinding.call(this);
        if (this.cardview) {
            this.cardview.onAfterDataBinding();
        }
    },
    onRedrawCore: function(){
        this.table.makeTitleRow();
    },
    dispose: function(){
        if (this.table) {
            document.controller.disposeObject(this.table);
            delete this.table;
        }
        ListBuilder.prototype.dispose.call(this);
    }
});
