// syracuse-ui/lib/desktop/field/collection/builder/listMenusController.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var ContextMenu = require('syracuse-ui/lib/desktop/field/menusController').ContextMenu;

var _$quickActions = {
    list: {
        array: ["$create", "$select", "$delete"]
    },
    record: {
        array: ["$create", "$edit", "$delete"]
    }
};

Object.keys(_$quickActions).forEach(function($type){
    var actions = _$quickActions[$type];
    actions.array.forEach(function($action){
        actions[$action] = true;
    });
});


function ListContextMenu(){
}

exports.ListContextMenu = helpers.defineClass(ListContextMenu, ContextMenu, {
    appendQuickActions: function($menuItem){
        var self = this;
        if (!self._quickActions) {
            self._quickActions = {};
            self.$quickActions.array.forEach(function($bind){
                var slot = document.createElement("div");
                slot.className = "s-list-quick-actions-slot";
                self._quickActions[$bind] = $(self._actionsSlot.insertBefore(slot, self._menusBtnSlot));
            });
        }
        var menuItem = document.itemFactory.create({
            $bind: $menuItem.$bind,
            //   $skin: "-link",
            $category: $menuItem.$category || "link",
            $skin: "s-list-quick-actions-link",
            $icon: {
                $mode: "icon",
                $path: "list/s-actions-",
                $width: "20px",
                $height: "20px"
            }
            //,
            //  $icon: $icon || null
        }, self.field);
        menuItem.$$container = self._quickActions[$menuItem.$bind];
        return menuItem;
    },
    clearMenuItems: function(){
        var self = this;
        if (self._quickActions) {
            Object.keys(self._quickActions).forEach(function($bind){
                $(self._quickActions[$bind]).remove();
            });
            delete self._quickActions;
        }
        self._isMenuBtnVisible = false;
        self._menusBtnSlot.style.display = "none";
        if (self.hasColumns) {
            self.field.builder.table.setContextMenusCol(null, true);
        }
    },
    create: function(field){
        var self = this
        if (field.list) {
            self.$skinMenus = "s-list-record-menus";
            self.$skinActionSlot = field.list.$skin + "-record-actions";
            self.$quickActions = _$quickActions.record;
            self.hasColumns = false;
            switch (field.list.$item.$format) {
                case "vgrid":
                    self.popupContent = {
                        id: document.controller.generateUUID(),
                        $$item: $(field._contextMenusSlot)
                    };
                    break;
                case "grid":
                    self.hasColumns = true;
                    break;
            }
        }
        else {
            self.$skinMenus = "s-list-menus";
            self.$skinActionSlot = "s-list-actions";
            self.$quickActions = _$quickActions.list;
        }
        ContextMenu.prototype.create.call(this, field, {
            onBeforeAddMenuItem: function($menuItem){
                if (self.$quickActions[$menuItem.$bind]) {
                    if (self.hasColumns) {
                        self.field.builder.table.setContextMenusCol($menuItem.$bind);
                    }
                    return self.appendQuickActions($menuItem);
                }
                else {
                    if (!self._isMenuBtnVisible) {
                        self._menusBtnSlot.style.display = "";
                        self._isMenuBtnVisible = true;
                        if (self.hasColumns) {
                            self.field.builder.table.setContextMenusCol($menuItem.$bind);
                        }
                        if ($menuItem.$bind == "$lazyload") {
                            return false;
                        }
                    }
                }
            }
        });
    },
    appendOpener: function(){
        this._actionsSlot = document.createElement("div");
        this._actionsSlot.className = this.$skinActionSlot;
        this.field._contextMenusSlot.appendChild(this._actionsSlot);
        this._menusBtn = document.createElement("a");
        this._menusBtn.className = "s-field-menus-btn s-close";
        
        this._isMenuBtnVisible = false;
        this._menusBtn.setAttribute("id", this.field.id + "-menu-btn");
        (this._menusBtnSlot = document.createElement("div")).className = "s-list-quick-actions-slot";
        this._menusBtnSlot.style.display = "none";
        return this._actionsSlot.appendChild(this._menusBtnSlot).appendChild(this._menusBtn);
    },
    removeOpener: function(){
        if (this._menusBtn) {
            document.site.emptyDom(this.field._contextMenusSlot);
            delete this._menusBtn;
            delete this._quickActions;
        }
    }
});
