// syracuse-ui/lib/desktop/field/collection/builder/referenceRecord.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Article = require("syracuse-ui/lib/common/article/article").Article;

function ReferenceRecord(){
}

exports.ReferenceRecord = helpers.defineClass(ReferenceRecord, Article, {
    loadRecord: function(list, record, $recordIndex){
        this.$recordUuid = record.$uuid;
        this.$recordIndex = $recordIndex;
        this.$facet = list.$recordFacet;
        this.$isEditMode = list.$isEditMode;
        this.$prototype = {
            $: {
                refField: list.$prototype.$item
            }
        };
        document.itemFactory.initializeItem(this, {}, list);
        this.loadBox({
            $uuid: record.$uuid,
            refField: record
        });
    },
    remove: function(){
        this.$$dataRow.remove();
    },
    drawBox: function(){
        var self = this;
        var list = self.boxParent;
        
        var div = document.createElement("div");
        div.className = list.$skin + "-record " + list.$recordCss;
        div.setAttribute("data-s-record", self.$recordUuid);
        self.$$dataRow = $(div);
        
        var div = document.createElement("div");
        div.className = list.$skin + "-record-data";
        var $$data = $(div).appendTo(self.$$dataRow);
        if (this.$isEditMode) {
            var btn = document.createElement("a");
            btn.className = list.$skin + "-record-del";
            self.$$del = $(btn).appendTo(self.$$dataRow).bind("click", function(){
                document.controller.notifyChange(list.getArticleParent(), list.$item.$bind, [{
                    $uuid: self.$recordUuid,
                    $index: self.$recordIndex,
                    $isDeleted: true
                }]);
                return false;
            });
        }
        document.itemFactory.load($$data, {
            $isTitleRowHidden: true,
            $bind: "refField",
            $isEditMode: false,
            $skin: "s-references-item",
            $inplace: true
        }, this);
        self.$$dataRow.appendTo(list._$$body);
    },
    dispose: function(){
        if (this.$$del) {
            this.$$del.unbind();
        }
        Article.prototype.dispose.call(this);
    }
});
