// etna/lib/x3js/collection-proxy.ts
"use strict";
/// !doc
/// [x3js dev guide](x3js-development-guide) > [x3js APIs](x3js-apis) > x3js collection
///
/// # x3js collection
///
/// A collection holds a list of child instances, for example the lines of a sales order instance.
///
/// Collections are not created independently, they belong their parent instance.
/// They are accessed like instance properties, for example:
///
/// ``` javascript
/// const myLines = mySalesOrder.lines$`;
/// if (myLines.length > 0) { ... }
/// ```

const util = require('etna/lib/supervisor/util');
const x3jsUtil = require('etna/lib/x3js/util');
const instanceProxy = require('./instance-proxy');
const constants = require("etna/lib/supervisor/constants");
import * as ez from 'ez-streams';

function path(col) {
	return util.path(col.$parent, col.collection.jsName);
}

const methods = {
	///
	/// #### Direct access to child instances
	///
	/// * `coln.length`: the length of the collection.
	// implemented in proxy.

	/// * `coln[i]`: the ith element of the collection, an instance. 
	///   `i` must vary between 0 and `coln.length - 1`. 
	///   `null` is returned if `i` is out of range.
	// implemented in proxy.

	///
	/// #### Adding and removing elements
	///
	/// * `coln.add(_, data = {}, i = null)`: adds an child instance to `coln`. 
	///   `data` is the initial data for the child instance. 
	///   `i` is the index at which the child will be added.
	///   If omitted (or `null`), the child is added at the end. 
	///   Returns the new child instance which has been created and added to the collection.
	add(_, data, i) {
		const x3Pos = this.add(_, i == null ? this.lines.length + 1 : i + 1);
		const line = this.index(_, x3Pos);
		Object.keys(this.collection.class.properties).forEach_(_, (_, k) => {
			if (data[k] !== undefined) line.set(_, k, data[k]);
		});
		return line;
	},
	/// * `coln.remove(_, i)`: removes the child at index `i`.
	remove(_, i) {
		this.del(_, i + 1);
	},
	/// * `coln.removeAll(_)`: removes all elements from the collection.
	removeAll(_) {
		this.empty(_);
	},

	///
	/// #### Reader and conversion to array
	///
	/// Collections implement the ez-streams reader API. 
	/// You can use this API to filter, transform, pipe, iterate, etc. 
	/// For example:
	///
	/// ``` javascript
	/// mySalesOrder.lines$
	/// .filter((_, line) => line.qty$ >= 10)
	/// .forEach(_, (_, line) => {
	///   // do something with line
	/// });
	/// ```
	/// See https://github.com/Sage/ez-streams/blob/master/README.md for details.
	///
	/// Note: you can use this API to convert the collection to a JavaScript array:
	///
	/// ``` javascript
	/// var linesArray = mySalesOrder.lines$.toArray(_);
	/// ```
	// this one called implicitly by proxy -- do not document
	reader() {
		var i = 0;
		return ez.devices.generic.reader(_ => i < this.lines.length ? instanceProxy.create(this.index(_, ++i)) : undefined, _ => i = 0);
	},

	///
	/// #### Collection level error, warning and info messages
	///
	/// * `coln.error(_, text)`: adds error `text` to `coln`.
	error(_, text) {
		this.collection.class.ASETERROR(_, path(this), text, constants.CST_AERROR);
	},
	/// * `coln.warn(_, text)`: adds warning `text` to `coln`.
	warn(_, text) {
		this.collection.class.ASETERROR(_, path(this), text, constants.CST_AWARNING);
	},
	/// * `coln.info(_, text)`: adds info `text` to `coln`.
	info(_, text) {
		this.collection.class.ASETERROR(_, path(this), text, constants.CST_AINFO);
	}
};

exports.create = collection => {
	if (!collection) return null;
	return new Proxy(collection, {
		get(collection, name) {
			if (name === 'length') {
				return collection.lines.length;
			} else if (methods[name]) {
				return methods[name].bind(collection);
			} else {
				var i = parseInt(name);
				if (!isNaN(i)) return instanceProxy.create(collection.lines[i]);
				else if (/^(inspect|toString|constructor)$/.test(name)) return collection[name] && collection[name].bind(collection);
				else if (ez.devices.generic.empty.reader[name]) // make it
					return ez.devices.generic.empty.reader[name].bind(methods.reader.call(collection));else throw new Error("invalid collection member: " + name);
			}
		}
	});
};
