// etna/lib/x3js/enum-proxy.ts
/// !doc
/// [x3js dev guide](x3js-development-guide) > [x3js APIs](x3js-apis) > x3js enums
///
/// # x3js enums
///
/// Enums are lists of integer constants.
///
/// Enums values have localized texts which are managed in the database (SQL for now, mongo later).
///
/// #### Defining an enum
///
/// An enum is defined with the `x3js.defineEnum(...)` function:
///
/// ``` javascript
/// import * as x3js from 'x3js';
///
/// export default x3js.defineEnum({
/// 	name: 'HighLow',
/// 	dictId: 6,
/// 	values: {
/// 		HIGH: 1,
/// 		LOW: 2,
/// 	}
/// });
/// ```
///
/// #### Using an enum
///
/// ``` javascript
/// import HighLow from '../../system/enums/high-low');
///
/// // assigning one of its values
/// obj.level$ = HighLow.HIGH;
/// ```
///
/// Note; an exception will be thrown if you mistype the enum code (`HighLow.HIGGH` for example).

require('harmony-reflect'); // to get ES6 Proxy API
import { _ } from 'streamline-runtime';
import * as flowControl from '../engine/runtime/flowControl';
const glob = require('streamline-runtime').globals;
import * as x3jsUtil from './util';

type Dict<T> = { [name: string]: T; };

export interface EnumDef {
	module: string;
	name: string;
	values: Dict<number>;
	dictId: number;
}

class EnumHandler {
	def: EnumDef;

	constructor(def: EnumDef) {
		this.def = def;
	}
	get $proxyType() { return 'enum'; }
	///
	/// #### Retrieving member names
	///
	/// * `MyEnum.$name(val)`: reverse lookup 
	///   Returns the name given an enum value.
	$name(val: number) {
		const keys = Object.keys(this.def.values);
		// fast return when enum has all values between 1 and max

		if (this.def.values[keys[val - 1]] === val) return keys[val - 1];
		// slower lookup
		for (var i = 0; i < keys.length; i++) {
			if (this.def.values[keys[i]] === val) return keys[i];
		}
		return null;
	}
	/// * `MyEnum.$names()`: returns the list of enum names, as an array of strings.
	$names() {
		return Object.keys(this.def.values);
	}
	/// * `MyEnum.$values()`: returns the list of enum values, as an array of integers.
	$values() {
		return Object.keys(this.def.values).map(name => this.def.values[name]);
	}
	///
	/// #### Retrieving localized texts
	///
	/// * `MyEnum.$text(_, val)`: returns localized text for the enum value.
	$text(val: number) {
		return x3jsUtil.wait(_ => {
			const text = flowControl.functions.MESS.fn(_, val, this.def.dictId, 1);
			if (text == null) throw new Error(`${ this.def.name }: invalid id, text is null`);
			return text;
		});
	}
}

const proxies: Dict<any> = {};

export function define(def: EnumDef) {
	if (!def.module) throw new Error(`enum module missing`);
	if (!def.name) throw new Error(`enum name missing`);
	//if (!def.dictId) throw new Error(`${def.name}: dict code missing`);
	if (!def.values) throw new Error(`${ def.name }: values missing`);
	var key = def.module + '.' + def.name;
	return proxies[key] || (proxies[key] = new Proxy(new EnumHandler(def), {
		get(handler: EnumHandler, name: string) {
			if (name[0] === '$') {
				const member = x3jsUtil.any(handler)[name];
				if (!member) throw new Error(`Enum ${ def.name }: invalid member: ${ name }`);
				return typeof member === 'function' ? member.bind(handler) : member;
			}
			const val = handler.def.values[name];
			// see later about throwing when value not found.
			return val;
			/*
			if (val == null) {
				return null;
				if (typeof name === 'symbol') return val;
				if (name === 'inspect') return null;
				if (name === 'toString') return () => "enum " + def.name;
				throw new Error(`invalid enum value: ${ name }, expected one of ${ Object.keys(handler.def.values).join(', ') }`);
			}
			return val;*/
		},
		ownKeys(handler) {
			return Object.keys(handler.$values);
		}		
	}));
};
