// syracuse-tablet/html/js/wc/authoring/authoringCommons.js
"use strict";

var modules = require('syracuse-tablet/html/js/wc/common/modules');

exports.getCurrentUserLang = function() {
	var up = modules.get("appController").App.getCurrentUserProfile();
	return up.selectedLocale.code;
}

exports.getCurrentPage = function() {
	var navStack = modules.get("navStack").getMainStack();
	var currentPage = navStack && navStack.getTopPage();
	return currentPage;
}

exports.getCurrentUserUuid = function() {
	var up = modules.get("appController").App.getCurrentUserProfile();
	return up.$uuid;
}

exports.setAuthoringActive = function(active) {
	modules.get("appController").App.setAuthoringActive(active);
}

exports.queryLocalePreferences = function() {
	return modules.get("ajax").request("/sdata/syracuse/collaboration/syracuse/localePreferences?representation=localePreference.$bulk")
		.then(function(result) {
			return result.data && result.data.$resources || [];
		});
}

exports.getSecurityProfile = function() {
	var up = modules.get("appController").App.getCurrentUserProfile();
	return up && up.securityProfile;
}
