// etna/lib/supervisor/constants.ts
"use strict";

export const CST_ADEL = 16;
export const CST_AERROR = 4;
export const CST_AFATAL = 5;
export const CST_AFIRSTPOS = 0;
export const CST_AINFO = 2;
export const CST_ALASTPOS = -1;
export const CST_ALL = 0;
export const CST_ANEW = 2;
export const CST_ANEWDEL = 1;
export const CST_ANO = 1;
export const CST_ANOTDEFINED = -1;
export const CST_AOK = 0;
export const CST_ASUCCESS = 1;
export const CST_ATYPBLOB = 9;
export const CST_ATYPCHAR = 7;
export const CST_ATYPCLOB = 10;
export const CST_ATYPDATE = 8;
export const CST_ATYPDATETIME = 12;
export const CST_ATYPDECIMAL = 4;
export const CST_ATYPDOUBLE = 6;
export const CST_ATYPFLOAT = 5;
export const CST_ATYPINSTANCE = 13;
export const CST_ATYPINTEGER = 3;
export const CST_ATYPSHORTINT = 2;
export const CST_ATYPTINYINT = 1;
export const CST_ATYPUUID = 11;
export const CST_AUPD = 4;
export const CST_AWARNING = 3;
export const CST_AYES = 2;
export const CST_ALEVFOLD = 1;
export const CST_ALEVCPY = 2;
export const CST_ALEVFCY = 3;
export const CST_ALEVUSR = 4;
export const CST_ALEVLEG = 5;
export const CST_ALOCK = 1;
export const CST_AOUTSEARCH = 2;
export const CST_ADUPKEY = 3;
export const CST_AOUTKEYS = 4;
export const CST_ANOREC = 5;
export const CST_ARECTICKUPD = 6;
export const CST_ARECTICKDEL = 7;
export const CST_ALLERRORS = '+';
export const CST_ACURRENT = 1;
export const CST_C_NAME_CLASS_CONTEXT = 'C_ACTX';
