// etna/lib/supervisor/gateway.ts
import { _ } from 'streamline-runtime';
import * as util from './util';
import * as runtime from '../engine/runtime/runtime';
import * as etnaClient from '../supervisor/client';
import { AGATEWAYADX } from '../supervisor/builtins/AGATEWAYADX';
const trace = require("syracuse-trace/lib/helper").getTracer("etna.supervisor");

export const $exported = true;

export function gateway(_: _,  script_name: string, operation_name: string, payload: any) {
	const context = util.currentContext();
	const endpoint = require("syracuse-collaboration/lib/helpers").AdminHelper.getEndpoint(_, {
		dataset: util.currentContext().endpointDataset
	});

	const client = etnaClient.getClient(_, context.session, endpoint);
	const resp = client.jsonSend(_, {
		head: {
			url: "/sdata/x3/erp/" + util.currentContext().endpointDataset + "/$service/rpc?module=" + script_name + "&name=" + operation_name,
			method: "POST"
		},
		payload: payload
	});

	trace.debug && trace.debug("gateway response:" + resp.body);
	return resp.body;
};

export function rpc(_: _,  context: any) {
	const body = context.request.readAll(_);

	trace.debug && trace.debug("rpc module:" + context.qs.module);
	trace.debug && trace.debug("rpc name: " + context.qs.name);
	trace.debug && trace.debug("rpc body: " + body);

	try {
		var _this = new AGATEWAYADX(context.superv, body).init(_);

		var module = runtime.requireScript(_, context.qs.module);
		var args = {
			THIS: {
				type: "LY",
				value: _this
			},
			FUNCTION: {
				type: "AS" + context.qs.name.length,
				value: context.qs.name
			}
		};
		runtime.executeProg(_, module, "$GATEWAY", args);
		context.response.end(_this.stringify(_));
	} catch (e) {
		util.traceException(e);
	}
};
