// etna/lib/supervisor/slotBlob.ts
"use strict";

const glob = require('streamline/lib/globals');
const ez = require('ez-streams');
const SlotLob = require("etna/lib/supervisor/slotLob").SlotLob;
const tblbfile = require('etna/lib/engine/runtime/tblbfile');
const constants = require("etna/lib/supervisor/constants");
const Blbfile = tblbfile.Blbfile;

export class SlotBlob extends SlotLob {
	constructor(parent, property) {
		super(parent, property);
	}

	set(_, value, raw) {
		if (value !== undefined && typeof value === 'string') {
			//dbg:console.log(this.name+" =>Blbfile");
			value = tblbfile.fromBase64(value);
		}
		super.set(_, value, raw);
		this.$wasUpdated = true;
	}

	get(_, raw) {
		if (raw) return this._value = this._value || new Blbfile();

		if (!this._value) {
			var data = this.read(_);
			this._value = data || new Blbfile();
		}
		return super.get(_, raw);
	}

	read(_) {
		if (this.property.type.readMedia) {
			return new Blbfile(this.property.type.readMedia(_, this));
		}
		return this.readRecord(_, "ABLOB", "BLOB");
	}

	insertSlotExtended(_) {
		if (!this.$wasUpdated || this.property.data.LOBTAB !== "ABLOB") return constants.CST_AOK;
		const keys = this.getKeys(_);

		return this.updateRecord(_, "ABLOB", "BLOB", {
			CODBLB: keys.CODBLB,
			IDENT1: keys.IDENT1,
			IDENT2: keys.IDENT2,
			IDENT3: keys.IDENT3,
			NAMBLB: ' ',
			TYPBLB: 0,
			CNTTYP: ' '
		}, ez.devices.buffer.reader);
	}

	updateSlotExtended(_) {
		return this.insertSlotExtended(_);
	}

	deleteSlotExtended(_) {
		const param = this.$parent.supervisor.sqlDriver.param;
		const keys = this.getKeys(_);
		//dbg:console.log("deleteLob name:"+slot.name+" keys:",keys);

		const params = [keys.CODBLB, keys.IDENT1, keys.IDENT2, keys.IDENT3];

		var i = 0;
		const sql = "delete From " + this.table + " Where" + " CODBLB_0=" + param(i++) + " and IDENT1_0=" + param(i++) + " and IDENT2_0=" + param(i++) + " and IDENT3_0=" + param(i++);

		return this.$parent.supervisor.executeSql(_, sql, params);
	}
}

;

exports.newSlotBlob = property => (parent => new SlotBlob(parent, property));
