// etna/lib/supervisor/statistics.ts
"use strict";
/*
http://localhost:8124/sdata/x3/erp/etnaOracle/STATS('DOM')?representation=STATS~DOM.$cube
*/

const glob = require('streamline/lib/globals');
import * as util from '../supervisor/util';
const HttpError = util.HttpError;
import * as bcd from '../engine/runtime/bcd';
import { Clbfile } from '../engine/runtime/tclbfile';
import * as runtime from '../engine/runtime/runtime';
import * as ASYR from '../supervisor/builtins/ASYR';
import * as ASYRSND from '../supervisor/builtins/ASYRSND';
import * as ASYRRCV from '../supervisor/builtins/ASYRRCV';
const tracerJs = require('syracuse-core').getTracer("etna.supervisor");

function loadScript(_: _,  script) {
	try {
		var mod = runtime.requireScript(_, "WMSTS" + script);
		mod.$4gl = true;
		return mod;
	} catch (e) {
		throw new HttpError(404, "statistic not found: " + script);
	}
}

export function proto(_: _,  httpContext, script, facet) {
	const mod = loadScript(_, script);
	const psyr = new ASYR.Constructor(httpContext);
	const args = {
		ACTX: {
			type: "AY",
			value: glob.context.x3session.actx
		},
		PSYR: {
			type: "AY",
			value: psyr
		},
		WCLOB: {
			type: "AT",
			value: new Clbfile("")
		},
		TIME_UPD: {
			type: "AN",
			value: bcd.fromDouble(0)
		},
		LEVEL_NUM: {
			type: "BS",
			value: 0
		},
		TYPE_REPRES: {
			type: "BS",
			value: "$query"
		}
	};

	runtime.executeProg(_, mod, "PROTO_JSON", args);
	return args.WCLOB.value.x3ToString();
};

export function query(_: _,  httpContext, script, query) {
	const mod = loadScript(_, script);
	const asnd = new ASYRSND.Constructor(httpContext);
	const arcv = new ASYRRCV.Constructor(httpContext.superv, httpContext.qs);
	const args = {
		ASND: {
			type: "LY",
			value: asnd
		},
		ARCV: {
			type: "LY",
			value: arcv
		}
	};

	runtime.executeProg(_, mod, "$GET", args);
	tracerJs.debug && tracerJs.debug(args.ASND.value.DATA);
	return args.ASND.value.DATA.toString("utf8");
};
