// syracuse-ui/lib/field/chart/flashWidget.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Article = require("syracuse-ui/lib/article/article").Article;
var ChartArticle = require("./chartArticle").ChartArticle;
var syraUtil = require('syracuse-ui/lib/fusion/core/client/sap/util').Syra;
var flashWrapper = require("./flashWrapper").flashWrapper;
var util = require('syracuse-ui/lib/fusion/tools/util');
var sapUtil = require('syracuse-ui/lib/fusion/core/client/sap/util').Fusion;

var debug = false;

function _debugSyraTrans() {
	document.syraTrans = document.syraTrans || {};
	document.syraTrans.logSwitches = {
		all: true
	};
}

function RecordArticle() {}

exports.RecordArticle = helpers.defineClass(RecordArticle, Article, {
	drawBox: function() {
		var slot = document.createElement("div");
		slot.style.display = "none";
		this.page.loadNewItem(this.boxParent.layoutSlot.appendChild(slot), {
			$category: "menus",
			$isMenusBag: true
		}, this);
	},
});

function FlashWidget() {}

exports.FlashWidget = helpers.defineClass(FlashWidget, ChartArticle, {
	_isRawFlash: function() {
		var type = syraUtil.getFieldType(this.attachedField || this);
		return type == syraUtil.dataTypes.flash;
	},
	_executeRecordMenu: function(context) {
		if (!this.recordArticle) {
			this.recordArticle = new RecordArticle();
			this.recordArticle.$prototype = this.$prototype.$item;
			this.page.initializeNewItem(this.recordArticle, {}, this);
			this.recordArticle.loadBox(context.$item);
		} else {
			this.recordArticle.applyChange(context.$item);
		}
		var $links = context["$links" || "$actions"];
		if ($links)
			this.recordArticle.menuItems[Object.keys($links)[0]][0].click();
	},
	showDiagnoses: function(exception, diagnoses, message) {
		var diag, page = syraUtil.getPage(this),
			ret = false;
		if (exception) {
			diag = [util.makeDiagnosisFromException(exception)];
		} else
		if (message) {
			diag = [util.makeDiagnosis(message, "warning")];
		}
		if (page && (diag || diagnoses)) {
			syraUtil.showDiagnoses(diag || diagnoses, page, null, false);
			ret = true;
		}
		return ret;
	},
	fxgetinitialdata: function(callback, myData) {
		callback(null, null, !this._isRawFlash() ? this.$prototype : null);
		this._protoUpToDate = true;
	},
	fxgetctxmenu: function(callback, myData, serieUuid, serieLineId) {
		syra_site.siteFunctions.onBeforeClick();
		if (!this.attachedField.$isDisabled) {
			this.page.externalAdapter.onGraphClickPicker({
				field: this,
				rcdUuid: serieUuid,
				rcdLineId: serieLineId,
				pickerType: "menus",
				grapLevel: "measure",
				doEvent: function(err, menus) {
					var errEx = null;
					if (!menus && !err) {
						// TODO : default error
						errEx = [{
							"$message": "not yet implemented",
							"$severity": "warning"
						}];
					}
					callback(errEx || err, menus);
				}
			});
		}
	},
	fxexeclink: function(callback, myData, link, options) {
		var data = {
			field: this,
			link: link
		};

		function _result(res) {
			debug && console.log("FlashWidget.fxexeclink._reply");
			var err = !res ? {
				"$message": "error"
			} : res.$diagnoses || null;
			// No reply is expected
			callback(err, null);
		}
		if (options && options.legacy) {
			debug && _debugSyraTrans();

			var args = link.linkCtx.split("=");
			// LINKVUE = SUBAPSMON:LINKVUE(FONCTION,OBJET,PARAM)
			this.page.externalAdapter.onBlockExRpc({
				field: this.attachedField,
				call: {
					proxy: "LINKVUE",
					values: [args[0].split(",")[0], args[1].substr(0, 3), args[2]],
					callback: _result
				}
			});

		} else
		if (this.page.externalAdapter.onGraphClickLink(data)) {
			this._executeRecordMenu(link);
		}
	},
	fxexecdrill: function(callback, myData, drillParams) {
		delete this._protoUpToDate;
		this.fetch({
			jsonParams: {
				$axes: drillParams.$axes,
				$slicer: drillParams.$slicer
			}
		});
	},
	fxnamedsvccall: function(callback, myData, svcParams) {
		alert("Hi, i'm \"" + myData.id + "\" and y should process named service \"" + svcParams.svcId + "\"");
		callback(null, null);
	},
	fxhelpflash: function(callback, myData, svcParams) {
		var helpLegacy = true;
		if (helpLegacy) {
			var page, helpUrl, session = syraUtil.getFusionController(this);
			if (svcParams.key && session && (session = session.getSession())) {
				page = syraUtil.getPage(this);
				helpUrl = session.getFlashHelpUrl(svcParams.key);
				if (page && helpUrl) {
					page.openHelp(null, helpUrl);
				}
			}
		} else {
			var item, hlpParams = {
					"type": (svcParams && svcParams.type) || sapUtil.helpHandler.vpd.cat
				};
			if (svcParams.key && (item = svcParams.key.lastIndexOf("/")) >= 0) {
				hlpParams.key = svcParams.key.substr(++item);
				this.fxhelp(null, myData, hlpParams);
			}
		}
		if (callback) {
			callback(null, null);
		}
	},
	fxhelp: function(callback, myData, svcParams) {
		var hlpParams = {
			"tokens": []
		}, $fusionController = syraUtil.getFusionController(this);
		if ($fusionController && svcParams && svcParams.key && svcParams.type) {
			hlpParams.tokens[0] = svcParams.type == "xhfunc" ? sapUtil.helpHandler.fct.cat : (svcParams.type == "xhfield" ? sapUtil.helpHandler.fld.cat : svcParams.type);
			hlpParams.tokens[1] = (svcParams.key.split("."))[0];
			$fusionController.displayBusinessHelp(hlpParams);
		}
		if (callback) {
			callback(null, null);
		}
	},
	fxspro: function(callback, myData, request) {
		var self = this;
		debug && console.log("FlashWidget.fxspro: request=" + request);

		function _result(res) {
			debug && console.log("FlashWidget.fxspro._reply");
			var err = !res ? {
				"$message": "error"
			} : res.$diagnoses || null,
				reply = {};
			if (!err) {
				reply.data = self.wrapper.rawData;
				reply.result = {
					"data": res.argsModified[0],
					"$diagnoses": {
						"$message": res.argsModified[1],
						"$severity": res.argsModified[2]
					},
					"status": res.argsModified[3]
				};
			}
			callback(err, reply);
		}
		debug && _debugSyraTrans();
		this.page.externalAdapter.onBlockExRpc({
			field: this.attachedField || this,
			call: {
				//AINTERFL:AINTERFL(CLOBD,MESSAG,GRAV,OK)
				proxy: "AINTERFL",
				values: [request],
				callback: _result,
				onError: function() {
					_result({
						"argsModified": ["", [""],
							[""], 1
						]
					});
				}
			}
		});
	},

	__drawBox: function() {
		this.wrapper = this._createWrapper();
		this.domItem = this.wrapper.$$elt[0];
		if (!this.$item.$isListEmbeded) {
			this.articleParent.bind(this, this.$item.$bind);
		}
		if (this.$field && this.$field.$type == syraUtil.dataTypes.flash) {
			this.domItem.className += " s-field-flash";
		}
		this.layoutSlot.className += " flash-slot";
	},

	__setDataBind: function(value, record, metaData) {
		var valueEx = null,
			type;
		debug && console.log("FlashWidget.setDataBind: value=" + value);
		this.currentValue = value;
		if (this.wrapper) {
			type = this.attachedField ? syraUtil.getFieldType(this.attachedField) : null;
			//valueEx = type && type == syraUtil.dataTypes.collectionType && this.attachedField.dataset && Array.isArray(this.attachedField.dataset) && this.attachedField.dataset.length > 0 ? this.attachedField.dataset : null;
			//this.wrapper.setData(valueEx || this.dataset, this._protoUpToDate ? null : this.$prototype, metaData);
			if ((value && value.length > 0) || !this._protoUpToDate || (metaData && !this._protoUpToDate)) {
				this.wrapper.setData(value, this._protoUpToDate ? null : this.$prototype, metaData);
			}
			this._protoUpToDate = true;
		}
	},
	__dispose: function() {
		this.domItem = this.wrapper = null;
	},
	__applyDesignMetaData: function(metadata, designing) {
		//
	},
	_createWrapper: function(componentType) {
		var id = this.id + "_flash",
			proto = this.$prototype,
			$fusionController = syraUtil.getFusionController(this);
		return new flashWrapper(componentType, {
			cont: $(this.layoutSlot),
			id: id,
			swf: "/syracuse-ui/addins/sageFlash/swf/FLASH",
			base: ".",
			bgcolor: "#FFFFFF",
			width: "100%",
			height: "100%"
		}, {
			instance: this,
			id: id
		}, {
			dataBinding: this.$item.$bind,
			legacyRequest: ($fusionController != null),
			legacyLinks: ($fusionController != null),
			userLang: $fusionController && $fusionController.getLang()
		});
	},
	getFieldEvalTitle: function() {
		return this.page.externalAdapter.getFieldEvalTitle(this);
	},
	getDataValue: function() {
		return this.wrapper.getData();
	},
	dirty: function() {
		return this.wrapper.isDirty();
	},
	focus: function() {
		this.domItem.focus();
		return true;
	},
	getEnumeration: function(code) {
		var $fusionController = syraUtil.getFusionController(this),
			fusionSess = $fusionController ? $fusionController.getSession() : null;
		return fusionSess && code ? fusionSess.getMenloc(code) : null;
	},
	onFieldInputEvent: function(event) {
		var self = this;
		if (self.page && self.page.externalAdapter) {
			self.page.externalAdapter.onFieldEvent({
				field: self,
				event: event,
				doEvent: function() {
					var fnct = "onInput" + event.type.charAt(0).toUpperCase() + event.type.slice(1);
					if (self[fnct]) {
						self[fnct](event.target, event);
					}
				}
			});
		}
	},
	applyShortCuts: function(shortcuts, event) {
		return true;
	}
});

function CubeChart() {}

exports.CubeChart = helpers.defineClass(CubeChart, FlashWidget, {});
