// syracuse-ui/lib/field/chart/flashHelpers.js
"use strict";
var jsxml = require("jsxml/lib/jsxml");
var helpers = require('syracuse-core/lib/helpers');
var forEachKey = helpers.object.forEachKey;

exports.xmlBuilder = function($type, options) {
	var types = {
		"application/x-agenda": ["AGD", "FAG"],
		"application/x-calendar": ["", ""],
		"application/x-gantt": ["GAT", "FGA"],
		"application/x-query-designer": ["", ""],
		"application/x-pad-player": ["", ""],
		"application/x-pad-designer": ["", ""],
	},
		root = {
			X3PORTAL: {
				Data: {},
				Technical: {
					HtcProvType: {
						$value: options && options.provider || "TABLEBLOC"
					},
					DataSrc: {
						$: {
							Type: options && options.type || types[$type][0],
							Code: ""
						}
					},
					VisCmp: {
						$: {
							Type: options && options.vtype || types[$type][1],
						},
					},
					UsedDataSrcPar: {},
					UserLang: {
						$value: options && options.userLang || "BRI"
					}
				},
				Messages: {}
			}
		};

	function _toXml() {
		return jsxml.stringify(root);
	}

	function _addNode(target, value, attr) {
		var parent = target.parent.node,
			node = {},
			key = target.key;
		return _addChild(parent, key, value, attr);
	}

	function _addChild(parent, key, value, attr) {
		var node = {};
		if (value != null) node.$value = value
		if (attr != null) node.$ = attr;
		return _appendNode(parent, key, node);
	}

	function _appendNode(parent, key, node) {
		if (parent[key] == null) {
			parent[key] = node;
		} else {
			if (!Array.isArray(parent[key])) {
				parent[key] = [parent[key]];
			}
			parent[key].push(node);
		}
		return node;
	}

	function _addPath(path, value, attr, opts) {
		var sep = opts && opts.sep || ".",
			parts = path.split(sep),
			target = parts.reduce(function(res, elem) {
				if (res.node == null) {
					res.node = res.parent.node[res.key] = {};
				}
				return {
					parent: res,
					key: elem,
					node: res.node[elem]
				};
			}, {
				parent: {
					parent: null,
					key: "root",
					node: root
				},
				key: "X3PORTAL",
				node: root.X3PORTAL
			});
		return _addNode(target, value, attr)
	}


	return {
		$: root.X3PORTAL,
		xml: _toXml,
		addNode: function(key, value, attr, opts) {
			return _addPath(key, value, attr, opts);
		},
		addChild: function(parent, key, value, attr) {
			if (typeof(key) === "object") {
				var keys = Object.keys(key).filter(function(k) {
					return k[0] !== "$";
				})
				if (keys.length !== 1) throw new Error("bad element expected 1 child got " + keys.length);
				return _appendNode(parent, keys[0], key[keys[0]]);
			}
			return _addChild(parent, key, value, attr);
		}
	};
};
