// syracuse-ui/lib/field/chart/highCharts.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var locale = require('syracuse-core/lib/locale');
var datetime = require('syracuse-core/lib/types/datetime');
var aspect = require('syracuse-core/lib/aspect');
var cubeHelper = require('syracuse-ui/lib/field/cube/cubeHelper');
var Article = require("syracuse-ui/lib/article/article").Article;
var TabularChartAuthoring = require("./authoring/tabularChartAuthoring").TabularChartAuthoring;
var httpHelper = require("syracuse-sdata/lib/httpHelper");
// require('syracuse-ui/deps/highcharts/highcharts');
require('syracuse-ui/deps/highcharts/highcharts.src');
require('syracuse-ui/deps/highcharts/modules/exporting');
Highcharts.theme = require('syracuse-ui/themes/desktop/sage/highcharts/theme').theme;

var highchartsOptions = Highcharts.setOptions(Highcharts.theme);

var forEachKey = helpers.object.forEachKey;
var objectCopy = helpers.object.copy;
var objectClone = helpers.object.clone;

var debug = false;
var inject = false;

var _localize = {};

var _style = (function() {
    var map = {
        point: "scatter",
        stick: "column"
    };

    return function($style) {
        return $style && (map[$style] || $style);
    };
})();

// Fusion page hacking for test purpose
inject && (function() {
    var WorkBook = require('syracuse-ui/lib/fusion/workBook').WorkBook;
    // var WinStack = require('syracuse-ui/lib/fusion/core/client/sap/window').WinStack;
    var propertyHook = {
        "application/x-array": function($itemPage, $facet, opts, prop, key) {
            if (key === 'BA') {
                // {
                //     "$type": "application/x-array",
                //     "$capabilities": "insert,delete",
                //     "$item": {
                //         "$properties": {
                //             "BA2": {
                //                 "$type": "application/x-string",
                //                 "$X3Name": "AQFF1_B1FLD1",
                //                 "$title": "{@S-AQFF1-3651}",
                //                 "$maxLength": 10,
                //                 "$groupName": "A"
                //             },
                //     }
                // }
                prop.$cube = {
                    "$mode": "table",
                    "$style": "xy",
                    "$displaysOneMeasure": false,
                    "$title": "{@S-AQFF1-3651}",
                    "$hierarchies": {
                        "BA2_h": {
                            "$title": "{@S-AQFF1-3651}",
                            "$properties": ["BA2"],
                            "$members": [],
                            "$isDefault": true
                        }
                    },
                    "$measures": {
                        "BA4": {
                            "$type": "application/x-decimal",
                            "$title": "{@S-AQFF1-23747}",
                            "$mode": "sum",
                            "$style": "stick",
                            "$isStacked": false,
                            "$isNormalized": false,
                            "$isDefault": true
                        },
                        "BA5": {
                            "$property": "BA5",
                            "$title": "{@S-AQFF1-23746}",
                            "$mode": "sum",
                            "$style": "stick",
                            "$isStacked": false,
                            "$isNormalized": false,
                            "$isDefault": true
                        },
                        "BA6": {
                            "$property": "BA6",
                            "$title": "{@S-AQFF1-23662}",
                            "$mode": "sum",
                            "$style": "stick",
                            "$isStacked": false,
                            "$isNormalized": false
                        },
                        "BA7": {
                            "$property": "BA7",
                            "$title": "{@S-AQFF1-23663}",
                            "$mode": "sum",
                            "$style": "stick",
                            "$isStacked": false,
                            "$isNormalized": false
                        }
                    }
                };
            }

            prop.$axes = [{
                "$title": "Articles vendus",
                "$hierarchies": [
                    ["BA2_h"]
                ]
            }, {
                "$title": "measures",
                "$hierarchies": [
                    ["$measures"]
                ]
            }];
        }
    };

    aspect.before(WorkBook, ["loadSheet"], function(context, args, result) {
        // loadSheet: function($itemPage, $facet, opts)
        var $itemPage = args[0] || {},
            $facet = args[1] || {},
            opts = args[2] || {};

        var $representation = $itemPage && $itemPage.$representation,
            $prototype = $representation && $representation.$prototype || {},
            $properties = $prototype.$properties;
        if ($representation.$url.match(/\.AQFF\./)) {
            helpers.object.forEachKey($properties, function(key, prop) {
                propertyHook[prop.$type] && propertyHook[prop.$type]($itemPage, $facet, opts, prop, key);
            });
        }
    });

    // aspect.before(WinStack, ["applyChange"], function(context, args, result) {
    //     // applyChange: function(winId, change, options)
    //     var winId = args[0] || {},
    //         change = args[1] || {},
    //         options = args[2] || {};
    //     var idEx = this._getWinIdEx(winId),
    //         model = this._windows[idEx.id],
    //         proto = model && model.getProto();
    //     helpers.object.forEachKey(change.entities, function(key, p) {
    //         var $type = proto && proto[key].$type;
    //         if ($type === "application/x-array") {
    //             // p.$style = ";font-family:MS Sans Serif;font-size:8pt;color:#000000;font-weight:normal;font-style:normal;background-color:#78A7F5";
    //         }
    //     });
    // });
})();

var _merge = function() {
        var args = arguments;
        return $.extend(true, null, args[0], args[1], args[2], args[3]);
    };

var _formatApi = require('syracuse-ui/lib/field/formatApi');

function _createChoices(title, items) {
    return {
        $type: "application/x-choice",
        $title: title,
        $value: {
            $type: "application/x-string",
            $enum: items.map(function(item) {
                return typeof(item) === "string" ? {
                    $value: item,
                    $title: _localize[item] || item
                } : {
                    $value: item.$value,
                    $title: item.$title || _localize[item.$value] || item.$value
                }
            })
        }
    };
};

// TODO: 
// - event management
// - dynamic loading of series

function RecordArticle() {};

exports.RecordArticle = helpers.defineClass(RecordArticle, Article, {
    drawBox: function() {
        this.page.loadNewItem($("<div style='display:none'/>").appendTo(this.boxParent.$$container), {
            $category: "links"
        }, this);
    },
});

//  "application/x-array"

function TabularChart() {}

exports.TabularChart = helpers.defineClass(TabularChart, Article, {
    applyDesignMetaData: function(metadata, onAuthoring) {
        var self = this;
        if (metadata.$style) {
            if (self._meta) {
                self._setStyle(metadata.$style);
            }
        }
    },
    _createChoices: function() {
        return _createChoices(_localize.chartType, ["area", "areaspline", "bar", "column", "line", "pie", "scatter", "spline"]);
    },
    getAuthoringWidget: function() {
        return new TabularChartAuthoring();
    },
    drawBox: function() {
        var self = this;
        this.$localization = this.getArticleParent().$prototype.$localization;
        _localize = locale.resources(module)();

        var id = self.id + "-chart";
        debug && console.log("HighCharts.drawBox " + self.id);

        $('<div />').attr('id', id).appendTo(self.$$container);

        var proto = self.$prototype;
        // create an alias
        proto.$properties = proto.$item.$properties;

        var options = _merge(self._getPref(), self._getOptions(id, proto));
        // cubeHelper._initializeMeta(self);
        self._initializeMeta();
        self._initializeToobar();
        // self.applyDesignMetaData(self.$item, false); //temp for  test
        self._chart = new Highcharts.Chart(options);
        this._showDrillUp(false);
        self.getArticleParent().bind(self, self.$item.$bind);
    },
    setDataBind: function(value, record, metaData) {
        debug && console.log("HighCharts.setDataBind " + this.id);
        //this.value = value;
        if (metaData) {
            this._processMeta(metaData);
        }
        if (value) {
            var self = this;
            var settings = value.reduce(function(res, current) {
                var categoryName = self._getMeasureLabel(current, self._meta.xAxis[0].code);
                categoryName && res.categories.push(categoryName);
                forEachKey(self._meta.series, function(key, s) {
                    var serie = res.series[key];
                    if (serie == null) {
                        serie = objectClone(s.options);
                        serie.data = [];
                        res.series[key] = serie;
                    }
                    serie.data.push(parseFloat(current[key]));
                });
                return res;
            }, {
                categories: [],
                series: {}
            });

            self._chart.xAxis[0].setCategories(settings.categories, false);

            // todo use series.setData
            // remove all current series
            var series = self._chart.series;
            for (var i = series.length; i > 0; i--) {
                // for (var i = 0; i < series.length; i++) {
                series[0].remove(false);
                // series[i].setData(settings.series[series[i].options.id.slice(3)].data);
            }
            // self._toolBox.removeAllSeries();
            //add series
            forEachKey(self._meta.series, function(key, s) {
                self._chart.addSeries(settings.series[key], false);
                // self._toolBox.addSeries({
                //     id: s.options.id,
                //     title: s.options.name
                // });
            });
            setTimeout(function() {
                self._chart.setSize(self.$$container.innerWidth(), self.$$container.innerHeight());
            }, 100);
        }
        this._showDrillUp(false);
    },

    _localizedText: function(val) {
        var $localization = this.$localization;
        return val && val.replace(/\{(@[\w-]+)\}/g, function(match, p1) {
            return ($localization && $localization[p1]) || match;
        });
    },

    _setSerie: function(id) {
        process.removeAllListeners(type);
        this._chart.series.map(function(s) {
            s.hide();
        });
        this._chart.get(id).show();
    },

    _setStyle: function(type) {
        var self = this;
        forEachKey(self._meta.series, function(key, s) {
            var series = self._chart.get(s.options.id);
            if (series) {
                self._setType(series, type);
            }
        });
        self._chart.redraw();
    },

    _getOptions: function(id, proto) {
        var self = this,
            cube = this.$prototype.$cube,
            defOptions = Highcharts.getOptions(),
            buttons = defOptions.exporting.buttons;

        var options = {
            chart: {
                renderTo: id,
                marginRight: 150
            },

            lang: {
                drillUp: _localize.drillUp
            },

            legend: {
                align: 'right',
                verticalAlign: 'top',
                x: 0,
                y: 130,
/*              align: 'center',
                 verticalAlign: 'bottom',
                 x: 0,
                 y: 0,*/
                // floating: true
            },

            title: (cube && self._localizedText(cube.$title)) || '',

            xAxis: [{
                categories: [],
                title: {
                    text: proto.$axes && proto.$axes.length > 0 && self._localizedText(proto.$axes[0].$title)
                }
            }],

            yAxis: [{
                min: 0,
                title: {
                    text: ''
                }
            }],

            plotOptions: {
                series: {
                    cursor: 'pointer',
                    point: {
                        events: {
                            click: function(event) {
                                // no longer drill up if shiftKey is pressed
                                if (!event.shiftKey) self._drill(this, event.shiftKey);
                            }
                        }
                    }
                }
            },

            tooltip: {
                formatter: function() {
                    var seriesMeta = self._meta.series,
                        measure = seriesMeta[this.series.options.id.slice(3)].measure,
                        label = '' + this.x + ': <b> ' + (measure ? measure.formatApi.format(this.y, measure.localeFormat) : this.y) + '</b>',
                        data = this.series.processedYData;
                    if (measure && (measure.$isNormalized) && data.length > 0) {
                        var sum = data.reduce(function(a, b) {
                            return a + b;
                        });
                        label += ' (' + measure.formatApi.format(this.y / sum * 100, measure.localeFormat) + '%)';
                    }
                    label += '<br/><span style="font-size: 10px; font-style:italic;">' + _localize.clickToDrill + '</span>'
                    // label += '<br/><span style="font-size: 10px; font-style:italic;">' + _localize.clickToDrill + '</span><br/><span style="font-size: 10px; font-style:italic;">Shift+Click to drill up</span>'
                    return label;
/*  var items = this.points || splat(this),
                     series = items[0].series,
                     s;
                     // build the header
                     s = [series.tooltipHeaderFormatter(items[0].key)];
                     // build the values
                     items.map(function(item) {
                     series = item.series;
                     s.push((series.tooltipFormatter && series.tooltipFormatter(item)) || item.point.tooltipFormatter(series.tooltipOptions.pointFormat));
                     });
                     // footer
                     s.push(options.footerFormat || '');
                     return s.join('');*/
                }
            },
            series: [],

            exporting: {
                buttons: {
                    drillUp: {
                        symbol: 'triangle',
                        x: -62,
                        symbolFill: 'rgba(250,250,250,0.5)',
                        hoverSymbolStroke: '#fff',
                        hoverSymbolFill: 'rgba(250,250,250,0.5)',
                        _id: 'drillUp',
                        _titleKey: 'drillUp',
                        onclick: function() {
                            self._drill(null, true);
                        }
                    },
                    printButton: buttons.printButton,
                    exportButton: buttons.exportButton
                }
            }
        };

        if (options.xAxis[0].title.text != options.title) {
            options.xAxis[0].title.text = options.title + ' / ' + options.xAxis[0].title.text;
        }

        return options;

    },

    _initializeMeta: function() {
        var proto = this.$prototype;
        var cube = proto.$cube;
        this._meta = {
            xAxis: [{}],
            series: {},
            displaysOneMeasure: cube && cube.$displaysOneMeasure,
            hasMeasureSelector: cube && cube.$hasMeasureSelector,
            hasStyleSelector: cube && cube.$hasStyleSelector,
            defaultStyle: _style(cube.$style) || 'column',
            drill: {
                minLevel: 1,
                maxLevel: 0,
                currLevel: 1,
                currAxisHier: 0,
                levelsProperties: [],
                members: [],
                stack: [{
                    $axes: proto.$axes,
                    $item: proto.$item,
                    $slicer: proto.$slicer
                }]
/*              code: proto.$codeStat || "UNKNOWN",
                 axis0FieldCode: self._meta.xAxis[0].code || ""*/
            }
        };
        this._processMeta(proto);
    },
    _initializeToobar: function() {
        var self = this;
        // self._toolBox && self._toolBox.show({
        //     series: (self._meta.hasMeasureSelector && self._meta.displaysOneMeasure),
        //     chartTypes: self._meta.hasStyleSelector
        // });
        // self._toolBox && self._toolBox.selectChartType(self._meta.defaultStyle);
    },
    _processMeta: function(proto) {
        this._processMeasures(proto);
        this._processLinks(proto);
        this._processAxes(proto);
    },
    _processMeasures: function(proto) {
        var self = this;
        var cube = this.$prototype.$cube;
        var items = this.$prototype.$properties;

        // Get measures definition. If measures are not defined in the second axis members (Cf. specs),
        // we used all the measures defined in the cube
        var measures = cube && cube.$measures;
        if (measures) {
            var restraintMeasures = measures;
            if (proto.$axes && proto.$axes.length > 1 && proto.$axes[1].$members && proto.$axes[1].$members[0][0].length > 0) {
                restraintMeasures = {};
                _.each(proto.$axes[1].$members[0][0], function(measureCode) {
                    restraintMeasures[measureCode] = measures[measureCode];
                });
            }
            _.each(restraintMeasures, function(m, key) {
                // Because duplicate properties is possible for cube definition... we keep the last defined
                var code = (m.$property || key);
                //options.xAxis[0].categories.push(measure.$title || proto.$properties[code].$title || "???");
                var measure = _merge(items[code], m);
                measure.formatApi = measure.formatApi || _formatApi.getApi(measure.$type);
                measure.formatObj = measure.formatObj || _formatApi.getLocale().getNumberFormatObj(measure.$type);
                measure.localeFormat = measure.localeFormat || measure.$format || measure.formatObj.numFormat;

                self._meta.series[code] = {
                    options: {
                        id: '#s-' + code,
                        // legendIndex: "",
                        name: self._localizedText(measure.$title) || '???',
                        stack: measure.$isStacked,
                        type: _style(measure.$style),
                        visible: !! (!self._meta.displaysOneMeasure || measure.$isDefault)
                    },
                    measure: measure
                };
                // self._chart && self._chart.addSeries(self._meta.series[code].options);
            });
        }
    },
    _processLinks: function(proto) {
        var self = this,
            reprUrl = self.page.$prototype && self.page.$prototype.$representationUrl;
        self._meta.links = [];
        proto.$links = proto.$links || {};

        if (reprUrl) {
            proto.$links = {};
            proto.$links.$pptslide = {
                "$title": _localize.powerpointSlide,
                "$url": reprUrl + "&pptMode=newSlide",
                "$type": httpHelper.mediaTypes.ppt_slide
            };
        }

        if (proto.$links) {
            _.each(proto.$links, function(link, id) {
                self._meta.links.push(self._getLinkDef("$links", id, link));
            }, self);
        }
        // need to validate the spec about these points
/*
         if(proto.$actions) {
         _.each(proto.$actions, function(link, id) {
         self._meta.links.push(self._getLinkDef("$actions", id, link));
         }, self);
         }
         if(self._meta.xAxis[0].code && proto.$properties && proto.$properties[self._meta.xAxis[0].code].$thumb && proto.$properties[self._meta.xAxis[0].code].$thumb.$links) {
         _.each(proto.$properties[self._meta.xAxis[0].code].$thumb.$links, function(link, id) {
         self._meta.links.push(self._getLinkDef("$links", id, link));
         }, self);
         }
         */
    },
    _processAxes: function(proto) {
        var cube = this.$prototype.$cube;
        if (!_isCubeValid(cube, proto)) return false;

        var self = this,
            currAxisHier = 0,
            axis0 = proto.$axes && proto.$axes[0];
        // TODO: review this
        // Get axis analysis
        var mainHLevelsTab = null;
        if (axis0) {
            _.each(axis0.$hierarchies, function(hierarchy, idx) {
                if (!mainHLevelsTab && hierarchy.length > 1 && hierarchy[1] == 1) {
                    mainHLevelsTab = cube.$hierarchies[hierarchy[0]].$properties;
                    if (mainHLevelsTab) {
                        self._meta.xAxis[0].code = mainHLevelsTab[hierarchy.length > 2 ? hierarchy[2] : 0];
                        currAxisHier = idx;
                    }
                }
            });
            if (!mainHLevelsTab && axis0.$hierarchies[0].length > 0) {
                mainHLevelsTab = cube.$hierarchies[axis0.$hierarchies[0][0]].$properties;
                if (mainHLevelsTab && mainHLevelsTab.length > 0) self._meta.xAxis[0].code = mainHLevelsTab[0];
            }
            // Get drill definition
            var properties;
            _.each(axis0.$hierarchies, function(hierarchie) {
                properties = cube.$hierarchies[hierarchie[0]].$properties;
                _.each(properties, function(property) {
                    this.levelsProperties[this.maxLevel++] = property;
                }, self._meta.drill);
            });
            if (axis0.$members) {
                _.each(axis0.$members, function(tuple) {
                    _.each(tuple, function(value) {
                        if (value.length > 0) {
                            this.members[this.currLevel++-1] = value[0];
                        }
                    }, this);
                }, self._meta.drill);
            }
        }
        self._meta.drill.code = proto.$codeStat || "UNKNOWN";
        self._meta.drill.axis0FieldCode = self._meta.xAxis[0].code || "";
        self._meta.drill.maxLevel = Math.max(self._meta.drill.maxLevel, self._meta.drill.minLevel);
    },

    // set graph style
    _setGraphStyle: function(cube, measures) {
        // Init
        var graphStyle = {},
            mDefault = null;
        var msCode = [],
            i = 0;
        var self = this;

        if (cube.$style == "pie") {
            _.each(measures, function(measure) {
/*              msCode.push(measure.property);
                 if(measure.definition.$isDefault)
                 mDefault = measure.property;
                 */
            });
        } else {
            _.each(measures, function(measure) {
/*
                 // So, "xy" style
                 switch (measure.definition.$style) {
                 case "stick":
                 break;
                 }
                 if(measure.definition.$isStacked) {
                 } else if(measure.definition.$isNormalized) {
                 }
                 if(measure.definition.$isDefault)
                 mDefault = measure.property;
                 msCode.push(measure.property);
                 */
            });
        }

        // Misc properties
        graphStyle.defaultMeasures = (!cube.$displaysOneMeasure ? msCode.join(";") : (mDefault ? mDefault : msCode[0]));
        graphStyle.hasMeasureSelector = !! cube.$hasMeasureSelector;
        graphStyle.hasStyleSelector = !! cube.$hasStyleSelector;

        return graphStyle;
    },
    _getLinkDef: function(type, id, link) {
        return {
            "type": type,
            "id": id,
            "link": link,
            "title": this._localizedText(link.$title)
        };
    },
    _getPref: function() {
        // TODO: change default preference by using authoring mode
        var options = Highcharts.getOptions();
        return _merge(options, {
            chart: {
                defaultSeriesType: 'column' // 'line', 'spline', 'area', 'areaspline', 'column', 'bar', 'pie', 'scatter',
            },
            credits: {
                enabled: false
            },
            legend: {
                layout: 'vertical',
                align: 'left',
                verticalAlign: 'top',
                x: 100,
                y: 70,
                floating: true
            },
            plotOptions: {
                column: {
                    pointPadding: 0.2,
                    borderWidth: 0,
                    //stacking : 'normal',
                }
            },
        });
    },
    _getMeasureLabel: function(measure, code) {
        // return cubeHelper.getMeasureLabel(measure, code, this.$prototype && this.$prototype.$properties);
        var label = this._localizedText(measure.$title);
        if (label) return label;
        var self = this;
        var item = self.$prototype.$properties[code];
        if (!item) return null;
        switch (item.$type) {
        case "application/x-choice":
            label = measure[code];
            item.$value.$enum.some(function(element, i, array) {
                if (element.$value === label) {
                    label = this._localizedText(element.$title);
                    return true;
                }
                return false;
            });
            break;
        case "application/x-date":
            item.formatApi = item.formatApi || _formatApi.getApi(item.$type);
            item.localeFormat = item.localeFormat || _formatApi.getLocale().getDateFormat(item.$format);
            label = datetime.parse(measure[code]).toString(item.localeFormat);
            break;
        case "application/x-integer":
        case "application/x-decimal":
            item.formatApi = item.formatApi || _formatApi.getApi(item.$type);
            item.formatObj = item.formatObj || _formatApi.getLocale().getNumberFormatObj(item.$type);
            item.localeFormat = item.localeFormat || item.$format || item.formatObj.numFormat;
            label = item.formatApi.format(measure[code], item.localeFormat);
            break;
        default:
            label = "" + measure[code];
            break;
        }
        return label;
    },

    // --------------------------------------------------------------------------------
    // Chart action management
    // --------------------------------------------------------------------------------
    _executeRecordMenu: function(context) {
        if (!this.recordArticle) {
            this.recordArticle = new RecordArticle();
            this.recordArticle.$prototype = this.$prototype.$item;
            this.page.initializeNewItem(this.recordArticle, {}, this);
            this.recordArticle.loadBox(context.$item);
        } else {
            this.recordArticle.applyChange(context.$item);
        }
        this.recordArticle.menuItems[Object.keys(context["$links" || "$actions"])[0]][0].click();
    },
    _execLink: function(context) {
        this._executeRecordMenu(context);
    },
    _showDrillUp: function(show) {
        // TODO: Does not work. Find an alternative
        // -----------------------------------------
        // var drillElt = this._chart.exportSVGElements && this._chart.exportSVGElements.filter(function(e) {
        //     return e.id === "drillUp";
        // });
        // if (!(drillElt[0] && drillElt[0].element)) return;
        // if (show) {
        //     drillElt[0].element.removeAttribute("display");
        // } else {
        //     drillElt[0].element.setAttribute("display", "none");
        // }
    },
    _drill: function(serie, up) {
        var self = this;
        var cube = this.$prototype.$cube;
        var meta = self._meta;
        var i = meta.drill.currAxisHier;
        var stack = self._meta.drill.stack;

        if (up) {
            // already on top ?
            if (stack.length < 2) return;
            stack.shift();
            meta.drill.currAxisHier--;
            self._showDrillUp(meta.drill.currAxisHier > 0);
            self.
            this._fetch(stack[0]);
            return;
        }

        var proto = stack[0],
            axis0 = proto.$axes && proto.$axes[0];

        if (!axis0) return;

        var members = axis0.$members = (axis0.$members || []);

        stack.unshift(objectClone(proto, true));
        proto = stack[0];
        axis0 = proto.$axes && proto.$axes[0];
        members = axis0.$members;
        var hierarchies = axis0.$hierarchies;

        var cubeHier = cube.$hierarchies[hierarchies[i][0]];
        if (hierarchies[i][2] == (cubeHier.$properties ? cubeHier.$properties.length - 1 : -1)) {
            hierarchies[i].splice(1, 2, 0, 0);
            hierarchies[i + 1].splice(1, 2, 1, 0);
            meta.drill.currAxisHier++;
        } else {
            hierarchies[i][2]++;
        }
        if (members.length < i + 1) {
            members[i] = [];
        }
        members[i].push([serie ? "" + serie.x : "???"]);

        self._showDrillUp(meta.drill.currAxisHier > 0);

        this._fetch(proto);
    },
    _fetch: function(proto) {
        this.fetch({
            jsonParams: {
                $axes: proto.$axes,
                $slicer: proto.$slicer
            }
        });
    },
    // --------------------------------------------------------------------------------
    // Chart management
    // --------------------------------------------------------------------------------
    _setChart: function(categories, series, data) {
        var self = this;
        alert('_setChart NIY');
/*
         self._chart.xAxis[0].setCategories(categories);
         self._chart.series[0].remove();
         self._chart.addSeries({
         name : name,
         proto : proto,
         //color : color || 'red'
         });
         */
    },
    _setType: function(series, newType) {
        var chart = series.chart;
        var newSeries = {
            id: series.options.id,
            type: newType,
            name: series.name,
            data: series.options.data,
            visible: series.visible,
            stack: series.stack,
            color: series.color
        };
        series.remove();
        chart.addSeries(newSeries, false);
    }
});

function _isCubeValid(cube, proto) {
    return !!(cube && cube.$hierarchies && proto.$axes && proto.$axes.length > 0 //
    && proto.$axes[0].$hierarchies && proto.$axes[0].$hierarchies.length > 0);
}

function _formatValue(v, format) {

}
