// syracuse-ui/lib/field/chart/orgChart.js
"use strict"; /* jshint node: true */
/* global Highcharts: true */
/* global alert: true */
var helpers = require('syracuse-core/lib/helpers');
var locale = require('syracuse-core/lib/locale');
var datetime = require('syracuse-core/lib/types/datetime');
var Article = require("syracuse-ui/lib/article/article").Article;
var httpHelper = require("syracuse-sdata/lib/httpHelper");
var TabularChartRecord = require('./tabularChartRecord').TabularChartRecord;
require('syracuse-ui/deps/highcharts-3.0.1/highcharts.src');
require('syracuse-ui/deps/highcharts-3.0.1/highcharts-more');
require('syracuse-ui/deps/highcharts-3.0.1/modules/exporting');
// Highcharts.theme = require('syracuse-ui/themes/desktop/sage/highcharts/theme').theme;
var debug = true;
// Highcharts.setOptions(Highcharts.theme);
var forEachKey = helpers.object.forEachKey,
    objectCopy = helpers.object.copy,
    objectClone = helpers.object.clone,
    stringify = helpers.object.stringify;

var _merge = Highcharts.merge;

var _formatApi = require('syracuse-ui/lib/field/formatApi');

function OrgChart() {}

exports.OrgChart = helpers.defineClass(OrgChart, Article, {
    onWindowResize: function(display) {
        debug && console.log("OrgChart.onWindowResize " + this.id + " : display=" + display + ", displayed=" + this.displayed + " , size=(" + this.layoutSlot.clientWidth + ", " + this.layoutSlot.clientHeight + ")");
        if (display) this.displayed = true;
        if (this.displayed && !this._chart && this.layoutSlot.clientWidth !== 0) {
            // defer chart creation when we are ready to display
            this._createChart();
        }
    },

    dispose: function() {
        this._chart && this._chart.destroy();
        this._localize = null;
        Article.prototype.dispose.call(this);
    },

    applyDesignMetaData: function(metadata, onAuthoring) {},

    drawBox: function() {
        this.dataset = null;
        this.$authoringType = "orgview$highCharts";
        this.$localization = this.getArticleParent().$prototype.$localization;
        this._localize = locale.resources(module)();
        // debug && console.log("HighCharts.drawBox " + this.id);
        var proto = this.$prototype;
        // create an alias
        proto.$properties = proto.$item.$properties;

        if (!this.$item.$isListEmbeded) {
            this.getArticleParent().bind(this, this.$item.$bind);
        }
    },

    setDataBind: function(dataRecordSet, parentDataRecord, metaData) {
        debug && console.log(Date.now() + " - OrgChart.setDataBind " + this.id + ": dataRecordSet=" + (dataRecordSet && dataRecordSet.length) + " , size=(" + this.layoutSlot.clientWidth + ", " + this.layoutSlot.clientHeight + ")");
        var self = this;
        if (metaData) {
            // this._processMeta(metaData);
        }
        this.dataset = dataRecordSet;

        var settings;
        if (this.dataset && this.dataset.length > 0) {
            // settings = this._settingsFromValue();
            settings = {};
            // var tm = "timer_createChart",
            //     fnc = function() {
            //         console.log(Date.now() + " - OrgCharts.delayed.setDataBind " + self.id + ": dataRecordSet=" + (dataRecordSet && dataRecordSet.length));
            //         if (!self._chart && self.layoutSlot.clientWidth !== 0) {
            //             return self._createChart({});
            //         }
            //         if (!self._chart) {
            //             setTimeout(fnc, 100);
            //         }
            //     };
            // if (this[tm]) clearTimeout(this[tm]);
            // this[tm] = setTimeout(fnc, 100);
            // ======
            // if (!this._chart && settings) {
            //     this._createChart(settings);
            // } else if (this._chart && settings) {
            //     var self = this;
        }
    },

    _buildTree: function(dataset) {
        var _map = this._map = {},
            $bindings = this.$prototype.$decorator.$bindings,
            binds = Object.keys($bindings).map(function(k) {
                return $bindings[k];
            });

        dataset.forEach(function(e, i, a) {
            var parentId = e[$bindings.$parent],
                id = e[$bindings.$id],
                parent = _map[parentId],
                node = _map[id];

            if (!parent && parentId.length !== 0) {
                parent = {
                    children: [],
                    id: parentId
                };
                _map[parentId] = parent;
            }
            node = node || {
                children: [],
                id: id
            };
            _map[id] = node;
            if (!node.parent) {
                node.parent = parent;
                node.desc = e[$bindings.$description];
                // node.desc ='<img class="" src="http://dummyimage.com/60x60/ff6600/ffffff"/>'
                // node.desc = '<img src="http://static.adzerk.net/Advertisers/bd294ce7ff4c43b6aad4aa4169fb819b.jpg" title="" alt="" border="0" height="50" width="50">' + //
                // '<span style="font-size: 9pt; font-style:italic;">' + e[$bindings.$description] + '</span>';
                // node.image = e[$bindings.$image];
                node.image = 'url(http://highcharts.com/demo/gfx/sun.png)';
                node.values = [];
                Object.keys(e).forEach(function(k) {
                    if (k.charAt(0) != "$" && binds.indexOf(k) == -1) {
                        node.values.push(e[k]);
                    }
                });
                parent && parent.children.push(node);
            }
            if (!parent) {
                _map[""] = node;
            }
        });
    },

    _createChart: function(settings) {
        var addEvent = Highcharts.addEvent,
            each = Highcharts.each;

        if (this.dataset && this.dataset.length > 0) {
            this._buildTree(this.dataset);
        }
        var self = this,
            // options = _merge(self._getPref(), self._getOptions(self.$prototype));
            options = {
                chart: {
                    renderTo: self.layoutSlot,
                    events: {
                        load: function() {
                            // _drawTest.call(this);
                            _drawOrg.call(this, self._map[""]);
                        }
                    }
                },
                title: {
                    text: 'Org chart view'
                },
                tooltip: {
                    formatter: function() {
                        return "tooltip";
                    }
                }
            };

        self._chart = new Highcharts.Chart(options);
    },

    _localizedText: function(val) {
        // TODO: use getLocalizeText
        var $localization = this.$localization;
        return val && val.replace(/\{(@[\w-]+)\}/g, function(match, p1) {
            return ($localization && $localization[p1]) || match;
        });
    },

    _redraw: function() {
        this._chart && this._chart.redraw();
    },

    _getFieldEvalTitle: function() {
        return this.page.externalAdapter.getFieldEvalTitle(this) || [];
    },

    // --------------------------------------------------------------------------------
    // Event management
    // --------------------------------------------------------------------------------
    _onMeasureClick: function(context) {},

    // --------------------------------------------------------------------------------
    // Chart management
    // --------------------------------------------------------------------------------
    destroy: function() {
        this._chart && this._chart.destroy && this._chart.destroy();
    }
});

// function label(str, x, y, shape, anchorX, anchorY, useHTML, baseline, className) {

function _label(ren, str, x, y) {
    var options = Highcharts.getOptions(),
        colors = options.colors;
    // ren.label("", -999, -999, null, null, null, true, null, 'worker').attr({
    return ren.label(str, x, y, null, null, null).attr({
        r: 5,
        // y: 30,
        // height: 60,
        fill: colors[1]
    }).css({
        color: 'black',
        fontWeight: 'normal',
        'text-align': 'center'
    });
}

function _drawOrg(root) {
    var ren = this.renderer,
        options = Highcharts.getOptions(),
        colors = options.colors,
        minHeight = 50,
        minWidth = 50;

    var bbox = null,
        worker = _label(ren, "", -999, -999).add();

    function positionNode(node, left, top) {
        var x = 0,
            y = 0,
            vGap = 10,
            hGap = 10,
            cw = 0;
        node.rect = node.rect || {};
        bbox = worker.attr({
            text: node.desc,
            // symbol: node.image
        }).getBBox();

        node.rect.width = Math.max(Math.ceil(bbox.width), minWidth);
        node.rect.height = Math.ceil(bbox.height);
        // node.rect.height = node.rect.height || minHeight;
        // node.rect.width = node.rect.width || minWidth;
        if (node.children.length > 0) {
            x = left;
            // the top coordinate of children
            y = top + node.rect.height + vGap;
            node.children.forEach(function(child) {
                cw = positionNode(child, x, y);
                // the left coordinate of the next child
                x += cw + hGap;
                debug && console.log("--> cw=" + cw + ", x=" + x);
            });
            // center Node between Left and(x - GapH) with top coordinate at Top
            node.rect.x = (left + (x - hGap) - node.rect.width) / 2;
            node.rect.y = top;
            debug && console.log("OrgChart.positionNode: " + node.desc + "(" + node.rect.x + "," + node.rect.y + ")" + "(" + (node.rect.x + node.rect.width) + "," + (node.rect.y + node.rect.height) + ")");
            // width of tree rooted at Node
            return Math.max(node.rect.width, x - hGap - left);
        } else {
            // position Node with top-left coordinate at (Left, Top)
            node.rect.x = left;
            node.rect.y = top;
            debug && console.log("OrgChart.positionNode: " + node.desc + "(" + node.rect.x + "," + node.rect.y + ")" + "(" + (node.rect.x + node.rect.width) + "," + (node.rect.y + node.rect.height) + ")");
            return node.rect.width;
        }
    }

    function drawNode(node) {
        var rect = node.rect,
            parent = node.parent,
            pRect = parent ? parent.rect : rect,
            p1 = {
                x: rect.x + rect.width / 2,
                y: rect.y
            },
            p2 = {
                x: pRect.x + pRect.width / 2,
                y: pRect.y + pRect.height
            };
        // draw Node into the rectangle
        var indent = "";
        for (var p = parent; p; indent += "  ", p = p.parent);
        debug && console.log(indent + node.desc + "(" + rect.x + "," + rect.y + "," + rect.width + "," + rect.height + ")");

        if (parent) {
            ren.path(['M', p1.x, p1.y, 'L', p1.x, (p1.y + p2.y) / 2, 'L', p2.x, (p1.y + p2.y) / 2, 'L', p2.x, p2.y]).attr({
                'stroke-width': 2,
                stroke: colors[3]
            }).add();
        }
        // var label = ren.label(node.desc, rect.x, rect.y).attr({
        // var label = ren.label(node.desc, rect.x, rect.y, node.image, null, null, true).attr({
        // var label = ren.label(node.desc, rect.x, rect.y).attr({
        var label = _label(ren, node.desc, rect.x, rect.y).add();

        Highcharts.addEvent(label.element, 'mouseenter', function() {
            label.attr('fill', colors[2]);
            label.css({
                color: 'white'
            });
        });
        Highcharts.addEvent(label.element, 'mouseleave', function() {
            label.attr('fill', colors[1]);
            label.css({
                color: 'black'
            });
        });
        label.on('click', function() {
            alert("id: " + node.id + ", parent: " + (node.parent ? node.parent.desc : "") + //
            "\nrect: " + stringify(rect) + //
            "\nvalues: " + node.values.reduce(function(a, b) {
                return a + "," + b;
            }));
        });

        // draw children
        node.children.forEach(function(child) {
            drawNode(child);
        });
    }

    // root.rect = {
    //     width: 0,
    //     height: 0
    // };
    positionNode(root, 20, 80);
    drawNode(root);
}

function _drawTest() {
    // Draw the flow chart
    var ren = this.renderer,
        colors = Highcharts.getOptions().colors,
        rightArrow = ['M', 0, 0, 'L', 100, 0, 'L', 95, 5, 'M', 100, 0, 'L', 95, -5],
        leftArrow = ['M', 100, 0, 'L', 0, 0, 'L', 5, 5, 'M', 0, 0, 'L', 5, -5];


    // Separator, client from service
    ren.path(['M', 120, 40, 'L', 120, 330]).attr({
        'stroke-width': 2,
        stroke: 'silver',
        dashstyle: 'dash'
    }).add();

    // Separator, CLI from service
    ren.path(['M', 420, 40, 'L', 420, 330]).attr({
        'stroke-width': 2,
        stroke: 'silver',
        dashstyle: 'dash'
    }).add();

    // Headers
    ren.label('Web client', 20, 40).css({
        fontWeight: 'bold'
    }).add();
    ren.label('Web service / CLI', 220, 40).css({
        fontWeight: 'bold'
    }).add();
    ren.label('Command line client', 440, 40).css({
        fontWeight: 'bold'
    }).add();

    // SaaS client label
    ren.label('SaaS client<br/>(browser or<br/>script)', 10, 82).attr({
        fill: colors[0],
        stroke: 'white',
        'stroke-width': 2,
        padding: 5,
        r: 5
    }).css({
        color: 'white'
    }).add().shadow(true);

    // Arrow from SaaS client to Phantom JS
    ren.path(rightArrow).attr({
        'stroke-width': 2,
        stroke: colors[3]
    }).translate(95, 95).add();

    ren.label('POST options in JSON', 90, 75).css({
        fontSize: '10px',
        color: colors[3]
    }).add();

    ren.label('PhantomJS', 210, 82).attr({
        r: 5,
        width: 100,
        fill: colors[1]
    }).css({
        color: 'white',
        fontWeight: 'bold'
    }).add();

    // Arrow from Phantom JS to Batik
    ren.path(['M', 250, 110, 'L', 250, 185, 'L', 245, 180, 'M', 250, 185, 'L', 255, 180]).attr({
        'stroke-width': 2,
        stroke: colors[3]
    }).add();

    ren.label('SVG', 255, 120).css({
        color: colors[3],
        fontSize: '10px'
    }).add();

    ren.label('Batik', 210, 200).attr({
        r: 5,
        width: 100,
        fill: colors[1]
    }).css({
        color: 'white',
        fontWeight: 'bold'
    }).add();
    // Arrow from Batik to SaaS client
    ren.path(['M', 235, 185, 'L', 235, 155, 'C', 235, 130, 235, 130, 215, 130, 'L', 95, 130, 'L', 100, 125, 'M', 95, 130, 'L', 100, 135]).attr({
        'stroke-width': 2,
        stroke: colors[3]
    }).add();

    ren.label('Rasterized image', 100, 110).css({
        color: colors[3],
        fontSize: '10px'
    }).add();

    // Browser label
    ren.label('Browser<br/>running<br/>Highcharts', 10, 180).attr({
        fill: colors[0],
        stroke: 'white',
        'stroke-width': 2,
        padding: 5,
        r: 5
    }).css({
        color: 'white',
        width: '100px'
    }).add().shadow(true);


    // Arrow from Browser to Batik
    ren.path(rightArrow).attr({
        'stroke-width': 2,
        stroke: colors[1]
    }).translate(95, 205).add();

    ren.label('POST SVG', 110, 185).css({
        color: colors[1],
        fontSize: '10px'
    }).add();

    // Arrow from Batik to Browser
    ren.path(leftArrow).attr({
        'stroke-width': 2,
        stroke: colors[1]
    }).translate(95, 215).add();

    ren.label('Rasterized image', 100, 215).css({
        color: colors[1],
        fontSize: '10px'
    }).add();

    // Script label
    ren.label('Script', 450, 82).attr({
        fill: colors[2],
        stroke: 'white',
        'stroke-width': 2,
        padding: 5,
        r: 5
    }).css({
        color: 'white',
        width: '100px'
    }).add().shadow(true);

    // Arrow from Script to PhantomJS
    ren.path(leftArrow).attr({
        'stroke-width': 2,
        stroke: colors[2]
    }).translate(330, 90).add();

    ren.label('Command', 340, 70).css({
        color: colors[2],
        fontSize: '10px'
    }).add();

    // Arrow from PhantomJS to Script
    ren.path(rightArrow).attr({
        'stroke-width': 2,
        stroke: colors[2]
    }).translate(330, 100).add();

    ren.label('Rasterized image', 330, 100).css({
        color: colors[2],
        fontSize: '10px'
    }).add();
}
