// syracuse-ui/lib/field/chart/tabularChartRecord.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Article = require("syracuse-ui/lib/article/article").Article;
//var RecordArticle = require("syracuse-ui/lib/field/array/record/recordArticle").RecordArticle;

function TabularChartRecord(){
};

exports.TabularChartRecord = helpers.defineClass(TabularChartRecord, Article, {
    drawBox: function(){
        var self = this;
        self._menusBoxSlot = document.createElement("div");
        self._menusBoxSlot.style.position = "absolute";
        self._menusBoxSlot.style.display = "none";
        self._menusBox = self.page.loadNewItem(self._menusBoxSlot, {
            $category: "menus",
            $isMenusBag: true,
            $skin: "s-record-actions",
            $layout: {
                $items: [{
                    $isMenusBag: true,
                    $category: "menus",
                    $title: "-",
                    $isBoxCollapsable: true,
                    $isPopupContent: true,
                    $id: self.id + "-menu-picker",
                    $skin: "s-record-menus",
                    $itemSkin: "s-field-menus-link",
                    onEnsureVisibility: function(menuGroup, $menuItemBind, menuItem){
                        if ($menuItemBind == "$lazyload") {
                            menuItem.hide(true);
                        }
                    }
                }]
            },
            onEnsureVisibility: function(menuGroup, $menuItemBind, menuItem){
                if (!menuGroup.menuGroupRoot.$isPickerVisible) {
                    self._menusBoxSlot.style.display = "";
                    menuGroup.menuGroupRoot.$isPickerVisible = true;
                }
            }
        }, self);
        self._menusBox.contextField = self;
        self.menuPicker = self.idMap[self.id + "-menu-picker"];
    },
    onMenuClick: function(menuItem){
        return this._onGraphMenuClick(menuItem);
    },
    openMenuBox: function($links, event, onGraphMenuClick){
        this.applyChange($links);
        this._onGraphMenuClick = onGraphMenuClick;
        this._menusBoxSlot.style.top = event.pageY + "px";
        this._menusBoxSlot.style.left = event.pageX + "px";
        this.menuPicker.doOpenPicker();
        this.menuPicker.$$body[0].style.top = event.pageY + "px";
        this.menuPicker.$$body[0].style.left = event.pageX + "px";
    },
    dispose: function(){
        if (this._menusBox) {
            document.controller.disposeObject(this._menusBox);
        }
        this._onGraphMenuClick = this.menuPicker = this._menusBox = null;
        Article.prototype.dispose.call(this);
    },
});
