// syracuse-ui/lib/field/chart/tabularChartRecord.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
//var Article = require("syracuse-ui/lib/article/article").Article;
var RecordArticle = require("syracuse-ui/lib/field/array/record/recordArticle").RecordArticle;

function TabularChartRecord(){
};

/*exports.TabularChartRecord = helpers.defineClass(TabularChartRecord, Article, {
 ensureDataSet: function() {
 return this.dataset = this.articleParent.dataset[this.$serverIndex] || {};
 },
 drawBox: function() {
 this._contextMenusSlot = document.createElement("div");
 this._contextMenusSlot.style.display = "none";
 this.boxParent.$$container[0].appendChild(this._contextMenusSlot);
 this.page.loadNewItem($(this._contextMenusSlot), {
 $category: "links"
 }, this);
 },
 dispose: function() {
 delete this.page;
 if (this.contextMenu) {
 document.controller.disposeObject(this.contextMenu);
 delete this.contextMenu;
 }
 Article.prototype.dispose.call(this);
 }
 });
 */
exports.TabularChartRecord = helpers.defineClass(TabularChartRecord, RecordArticle, {
   /*ensureDataSet: function(){
        return this.dataset = this.articleParent.dataset[this.$serverIndex] || {};
    }*/
});
