// syracuse-ui/lib/field/array/gridBuilder/tools/colResizerDDAgent.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function ColResizerDDAgent() {}

exports.ColResizerDDAgent = helpers.defineClass(ColResizerDDAgent, null, {
	start: function(dropItem) {
		this.resizedCol = dropItem.resizedCol;
		this.builder = dropItem.resizer.builder;
		document.site.ddManager.ddAgent = this;
		delete this.rightResizedCol;
		var offset = this.resizedCol.titleCell.getBoundingClientRect();
		this.left = offset.left;
		this.top = offset.top;
		this.height = offset.height;
		this.right = offset.right;
		this.bottom = offset.bottom;
		this.startX = this.resizedCol.resizer.getBoundingClientRect().left;
		this.resizedCol.resizer.style.cursor = this.cursor = "e-resize";

		if (this.builder.scroller.isProportionalWidth) {
			var resizedColIndex;
			for (var ii = 0, jj = this.builder.orderCols.length; ii < jj; ii++) {
				var col = this.builder.orderCols[ii];
				if (col.realWidth !== undefined) {
					col.widthValue = col.realWidth;
				}
				if (col == this.resizedCol) {
					resizedColIndex = ii;
				}
			}
			if (resizedColIndex != undefined) {
				if ((resizedColIndex + 1) < this.builder.orderCols.length) {
					this.rightResizedCol = this.builder.orderCols[resizedColIndex + 1];
					this.rightResizedCol.widthValue = this.rightResizedCol.titleCell.clientWidth;
				}
			}
		}
	},
	onDragMouseMove: function(target, event) {
		var cursor = "not-allowed";
		if ((event.pageY >= (this.top - 20)) && (event.pageY <= (this.bottom + 20))) {
			var newWidth = this.resizedCol.widthValue + (event.pageX - this.startX);
			if (newWidth > 50) {
				this.resizedCol.groupHead.style.width = this.resizedCol.groupBody.style.width = (this.newWidth = newWidth) + "px";
				if (this.rightResizedCol) {
					this.rightResizedCol.groupHead.style.width = this.rightResizedCol.groupBody.style.width = (this.rightResizedCol.widthValue = this.rightResizedCol.realWidth - (event.pageX - this.startX)) + "px";
				}
				cursor = "e-resize";
			}
		}
		document.site.body.style.cursor = cursor;
	},
	onDragMouseUp: function(target, event) {
		this.resizedCol.widthValue = this.newWidth;
		this.builder.onColumnResized(this.resizedCol);
		this.builder.onWindowResize();
		document.site.body.style.cursor = "default";
	},
	dispose: function() {
		this.startX = this.resizedCol = this.builder = this.resizedCol = this.rightResizedCol = null;
	}
});
