// syracuse-ui/lib/field/array/gridBuilder/tools/colReorderDDAgent.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function ColReorderDDAgent() {}

exports.ColReorderDDAgent = helpers.defineClass(ColReorderDDAgent, null, {
	start: function(dropItem) {
		document.site.ddManager.ddAgent = this;
		this.reordeCol = dropItem.reordeCol;
		this.list = dropItem.list;
		var offset = this.reordeCol.titleCell.getBoundingClientRect();
		this.top = offset.top;
		this.left = offset.left;
		this.boundary = this.list.titleRow.getBoundingClientRect();
		this.info = document.createElement("div");
		this.cueTop = document.createElement("div");
		this.cueBottom = document.createElement("div");
		this.scrollBoundary = this.list.scroller.getScrollBoundary();
		this.info.className = "s-grid-drag-image";
		document.site.layoutSlot.appendChild(this.info);
		this.cueTop.style.display = "none";
		this.cueBottom.style.display = "none";
		this.cueTop.className = "s-grid-drop-cue-top";
		this.cueTop.style.top = (this.boundary.top - 9) + "px";
		this.cueBottom.style.top = this.boundary.bottom + "px";
		document.site.layoutSlot.appendChild(this.cueTop);
		this.cueBottom.className = "s-grid-drop-cue-bottom";
		document.site.layoutSlot.appendChild(this.cueBottom);
		this.info.textContent = this.reordeCol.title.textContent;
		this.list.isColReordering = true;
	},
	onColumnHeaderEnterLeave: function(col, onEnter) {
		if (onEnter) {
			this.colTarget = col;
			this.targetX = this.colTarget.titleCell.getBoundingClientRect().left - 4; //4 cue width /2
			if (this.list.orderCols[this.list.orderCols.length - 1] == this.colTarget) {
				this.lastColBoundaryX = this.targetX + this.colTarget.realWidth / 2;
			}
			this.cueTop.style.left = this.cueBottom.style.left = this.targetX + "px";
			this.cueTop.style.display = this.cueBottom.style.display = "";
			document.site.dom.toggleClass(this.info, "s-drag-ok", this.colTarget != null);
		} else {
			if (this.colTarget && this.colTarget == col) {
				this.colTarget = null;
				delete this.lastColBoundaryX;
				this.cueTop.style.display = this.cueBottom.style.display = "none";
			}
			document.site.dom.toggleClass(this.info, "s-drag-ok", this.colTarget != null);
		}
	},
	autoScroll: function(event) {
		var self = this;
		var step = 40,
			margin = 80;
		if (self.autoScrolTimeout) {
			clearTimeout(self.autoScrolTimeout);
		}
		self.autoScrolTimeout = setTimeout(function() {
			if (self.scrollBoundary) {
				if (event.pageY > (self.scrollBoundary.bottom - margin)) {
					var scrollTop = self.scrollBoundary.bodySlot.scrollTop;
					if (scrollTop < (self.scrollBoundary.scrollHeight - self.scrollBoundary.height)) {
						self.scrollBoundary.bodySlot.scrollTop = Math.min(scrollTop + step, self.scrollBoundary.scrollHeight - self.scrollBoundary.height);
					}
				} else {
					if (event.pageY < (self.scrollBoundary.top + margin)) {
						var scrollTop = self.scrollBoundary.bodySlot.scrollTop;
						self.scrollBoundary.bodySlot.scrollTop = Math.max(scrollTop - step, 0);
					}
				}
				if (event.pageX > (self.scrollBoundary.right - margin)) {
					var scrollLeft = self.scrollBoundary.hscroller.scrollLeft;
					if (scrollLeft < (self.scrollBoundary.scrollWidth - self.scrollBoundary.width)) {
						self.scrollBoundary.hscroller.scrollLeft = Math.min(scrollLeft + step, self.scrollBoundary.scrollWidth - self.scrollBoundary.width);
					}
				} else {
					if (event.pageX < (self.scrollBoundary.left + margin)) {
						var scrollLeft = self.scrollBoundary.hscroller.scrollLeft;
						self.scrollBoundary.hscroller.scrollLeft = Math.max(scrollLeft - step, 0);
					}
				}
			}
		}, 10);
	},
	onDragMouseMove: function(target, event) {
		this.autoScroll(event);
		var top = Math.max(event.pageY, this.boundary.top);
		var left = Math.max(event.pageX, this.boundary.left);
		var inBoundary = (event.pageX == left && event.pageY == top);
		this.info.style.top = top + 15 + "px";
		this.info.style.left = left + 15 + "px";
		if (this.lastColBoundaryX) {
			if (event.pageX > this.lastColBoundaryX) {
				if (!this.dropAtEnd) {
					this.cueBottom.style.left = this.cueTop.style.left = (this.targetX + this.colTarget.realWidth) + "px";
					this.dropAtEnd = true;
				}
			} else {
				if (this.dropAtEnd) {
					this.cueBottom.style.left = this.cueTop.style.left = this.targetX + "px";
					this.dropAtEnd = false;
				}
			}
		}
	},
	onDragMouseUp: function(target, event) {
		document.site.dom.removeChild(this.info);
		document.site.dom.removeChild(this.cueTop);
		document.site.dom.removeChild(this.cueBottom);
		if (this.colTarget) {
			this.list.reorderColumns(this.reordeCol, this.colTarget, this.dropAtEnd);
		}
		document.site.body.style.cursor = "default";
	},
	dispose: function() {
		if (this.autoScrolTimeout) {
			clearTimeout(this.autoScrolTimeout);
		}
		if (this.list) {
			delete this.list.isColReordering;
		}
		this.colTarget = this.boundary = this.autoScrolTimeout = this.reordeCol = this.list = this.info = this.cueTop = this.cueBottom = this.scrollBoundary = null;
	}
});
