// syracuse-ui/lib/field/array/gridBuilder/tools/quickEditBuilder.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function QuickEdit() {}

helpers.defineClass(QuickEdit, null, {
	load: function(record, $itemPage) {
		this.list = record.list;
		if (this.list) {
			this.record = record;
			var cardDecorator = this.list.builder.cardDecorator;
			cardDecorator && cardDecorator.isRowCardMode && cardDecorator.toggleRowCard(record, false);
			this.card = this.list.builder.createCardRow(record);
			$itemPage.layoutSlot = this.card.contentSlot;
			$itemPage.inlinePageHost = this;
			this.resize();
			this.page = syra_site.pageLoader.load($itemPage);
			syra_site.currentQuickEdit = this;
		}
	},
	onInlinePageResized: function() {
		this.resize();
	},
	resize: function() {
		//debugger;
		//this.card.contentSlot.style.width = this.list._core.clientWidth + "px";
	},
	dispose: function() {
		this.page && this.page.dispose();
		this.card && syra_site.dom.removeChild(this.card.row);
		syra_site.currentQuickEdit = this.list = this.card = this.record = this.page = null;
	},
	onPreInitializeInlinePage: function(inlinePage, $article) {
		var $items = [];
		var $complex = [];
		var $fields = inlinePage.$prototype.$properties;
		for (var ii = 0, jj = this.list.builder.allColumns.length; ii < jj; ii++) {
			var col = this.list.builder.allColumns[ii];
			if (col.$bind && !col.$isHidden) {
				var $field = $fields[col.$bind];
				switch ($field && $field.$type) {
					case "image":
						$complex.push({
							$bind: col.$bind
						});
						break;
					default:
						$items.push({
							$bind: col.$bind
						});
						break;
				}
			}
		}
		var colCount = Math.min(Math.ceil($items.length / 3), 3);
		var colIndex = 0,
			$cols = [];
		for (var ii = 0; ii < colCount; ii++) {
			$cols.push({
				$layoutType: "stack",
				$items: []
			});
		}
		for (var ii = 0, jj = $items.length; ii < jj; ii++) {
			$cols[colIndex++].$items.push($items[ii]);
			if (colIndex == colCount) {
				colIndex = 0;
			}
		}
		$article.$layout = {
			$items: [{
				$layoutType: "row",
				$items: $cols
			}]
		};
		if (this.list.$complexBinds) {
			for (var ii = 0, jj = this.list.$complexBinds.length; ii < jj; ii++) {
				$complex.push({
					$bind: this.list.$complexBinds[ii]
				});
			}
		}
		if ($complex.length) {
			colIndex = 0, $cols = [], colCount = ($complex.length >= 2) ? 2 : 1;
			for (var ii = 0; ii < colCount; ii++) {
				$cols.push({
					$layoutType: "stack",
					$items: []
				});
			}
			for (var ii = 0, jj = $complex.length; ii < jj; ii++) {
				$cols[colIndex++].$items.push($complex[ii]);
				if (colIndex == colCount) {
					colIndex = 0;
				}
			}
			$article.$layout.$items.push({
				$layoutType: "row",
				$items: $cols
			});
		}
	},
	onAfterActionMenuExecute: function(page, menuItem, $menu) {
		var record = this.record;
		switch (menuItem.$sourceBind) {
			case "$save":
				var $binds = Object.keys(record.$prototype.$properties);
				var $delta = {};
				for (var ii = 0, jj = $binds.length; ii < jj; ii++) {
					var $bind = $binds[ii];
					if (page.dataset[$bind] != undefined) {
						$delta[$bind] = page.dataset[$bind];
					}
				}
				syra_site.deltaManager.applyObjectDelta(record.page, record.dataset, $delta);
				record.applyChange($delta);
				setTimeout(function() {
					exports.dispose(record.list);
				}, 100);
				return false;
			case "$abort":
				setTimeout(function() {
					exports.dispose(record.list);
				}, 100);
				return false;
		}
		return true;
	},
	_confirmClose: function(onConfirmed) {
		var self = this;
		syra_diagnose.showBox({
			$title: "Question", //"syra_local.aw_updateMessageTitle,
			$message: "Continue and cancel the modifications?",
			$type: "question",
			$buttonsTitle: {
				no: "Stay on current quickEdit",
				yes: "Continue"
			},
			callback: function(response) {
				if (response.$clientId == "yes") {
					exports.toggle(self.record, true);
					onConfirmed && onConfirmed();
				}
			}
		});
	},
	needCloseConfirm: function(activePage, onConfirmed) {
		if (!activePage || (!activePage.isMessageBox && this.page != activePage)) {
			if (this.page.isWorkingCopyUpdated()) {
				this._confirmClose(onConfirmed);
				return true;
			}
		}
		return false;
	}
});

exports.toggle = function(record, confirmed) {
	if (record.list.quickEdit) {
		if (!confirmed && record.list.quickEdit.needCloseConfirm()) {
			return;
		}
		var load = record.list.quickEdit.record != record;
		exports.dispose(record.list);
		if (!load) {
			return;
		}
	}
	//trigger action    
	var $mn = record.$menus.$quickEdit = helpers.object.clone(record.$menus.$edit, true);
	$mn.$target = "inline";
	syra_menus.click.menuId(record, $mn.$bind = $mn.$sourceBind = "$quickEdit");
};


exports.loadInlinePage = function(record, $itemPage, options) {
	record.list.quickEdit = new QuickEdit();
	record.list.quickEdit.load(record, $itemPage);
};
exports.disposeRecord = function(record) {
	if (record.list.quickEdit && record.list.quickEdit.record == record) {
		exports.dispose(list);
	}
};

exports.dispose = function(list) {
	list.quickEdit && list.quickEdit.dispose();
	list.quickEdit = null;
};
