// syracuse-ui/lib/site/bookmarks/bookmarkItem/bookmarkMenuBlock.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var ChildField = require('syracuse-ui/lib/field/child/childField').ChildField;
var BookmarkMenuBlockRecord = require('./bookmarkMenuBlockRecord').BookmarkMenuBlockRecord;

function BookmarkMenuBlock() {}

exports.BookmarkMenuBlock = helpers.defineClass(BookmarkMenuBlock, ChildField, {
	dispose: function() {
		this.subModuleRecord = null;
		ChildField.prototype.dispose.call(this);
	},
	loadBox: function() {
		this.$skin = this.$item.$skin = "s-nav-admin-block";
		this.ChildRecordClass = BookmarkMenuBlockRecord;
		this.subModuleRecord = this.articleParent.list.subModuleRecord;
		ChildField.prototype.loadBox.call(this);
		this.articleParent.domItem.className += "  " + this.$skin + "-item";
		if (this.articleParent.reorderPicker) {
			this.articleParent.reorderPicker.className += " s-nav-block-reorder";
		}
		this.onReorderItem(this.articleParent);
		this.isMenuBlock = true;
	},
	onReorderItem: function(record) {},
	onItemEnterLeave: function() {
		//important for compatibility
	}
});
