// syracuse-ui/lib/site/bookmarks/bookmarkItem/bookmarkMenuBlockRecord.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var ChildFieldRecord = require('syracuse-ui/lib/field/child/childFieldRecord').ChildFieldRecord;

function BookmarkMenuBlockRecord() {}

exports.BookmarkMenuBlockRecord = helpers.defineClass(BookmarkMenuBlockRecord, ChildFieldRecord, {
	drawBox: function() {
		var $skin = "s-nav-admin-block";
		this.$expanded = true;
		this.domItem = this.list._core;
		this.blockCard = this.domItem.appendChild(document.createElement("div"));
		this.blockCard.className = this.list.$skin + "-record";
		this.blockCardHeader = this.blockCard.appendChild(document.createElement("div"));
		this.blockCardHeader.className = this.list.$skin + "-record-head";
		this.$prototype.$properties.title.$isMandatory = true;
		if (this.articleParent.articleParent.menusSlot) {
			this.articleParent.articleParent.menusSlot.className = $skin + "-menus";
		}

		var prevTitleField = this.titleField;
		this.titleField = this.page.loadNewItem(this.blockCardHeader, {
			$bind: "title",
			$isEditMode: true,
			$isAutoSizeDisabled: true,
			$isTitlePlaceHolder: true,
			$title: this.page.localize.nvpEnterMenuBlock,
			$contentEditable: true,
			$skin: $skin + "-title-field",
			$isMenusHidden: false,
			$isDetailLinkDisabled: true,
			$useLocalizePicker: true,
			$isTitleHidden: true
		}, this);
		if (!this.itemsField) {
			this.itemsField = this.page.loadNewItem(this.blockCard, {
				$bind: "items",
				$isTitleHidden: true
			}, this);
			this.itemsField.body.className += " " + $skin + "-items-field";
			this.itemsField.menusSlot.className = "s-nav-admin-items-menus-slot";
		} else {
			this.blockCard.insertBefore(this.titleField.domItem, this.itemsField.domItem);
		}
		var itemRecord = this.articleParent.articleParent;
		this.titleField._dataValue.appendChild(this.itemsField.menusSlot);
		this.addExpandPicker();
		if (itemRecord.reorderPicker) {
			this.expandPicker.parentNode.insertBefore(itemRecord.reorderPicker, this.expandPicker);
		}
		this.titleField._dataValue.appendChild(itemRecord.menusSlot);
		this.blockCardHeader.syrainout = this;
		this.blockCardHeader.setAttribute("data-s-inout", this.id);
		if (prevTitleField) {
			this.removeItem(prevTitleField, true, true);
		}
	},
	onLocalizeField: function(openerMenu, data, response, $url) {
		this.articleParent.subModuleRecord.applyChange(data, response, $url);
		return false;
	},
	addExpandPicker: function() {
		if (!this.expandPicker) {
			this.expandPicker = document.createElement("a");
			this.expandPicker.setAttribute("href", "#");
			this.expandPicker.setAttribute("data-s-picker", this.id);
			this.expandPicker.syraIsDesignEnable = true;
			this.expandPicker.syraBoxId = this.id;
			this.expandPicker.syraBoxPickerId = "expand";
			this.expandPicker.className = "s-nav-admin-block-expand";
			this.expandPicker.title = this.page.localize.box_collapse;
			if (this.$item.$css) {
				this.expandPicker.className += " " + this.$item.$css;
			}
			this.titleField.domItem.insertBefore(this.expandPicker, this.titleField.domItem.firstChild);
		}
		this.ensureExpandedState();
	},
	onClickPicker: function(picker) {
		if (picker == this.expandPicker) {
			this.expandMenuBlock(!this.$expanded);
			return false;
		}
		return true;
	},
	expandMenuBlock: function(expanded) {
		this.$expanded = expanded;
		this.ensureExpandedState();
	},
	ensureExpandedState: function() {
		this.expandPicker.title = this.$expanded ? this.page.localize.box_collapse : this.page.localize.box_expand;
		document.site.dom.toggleClass(this.expandPicker, "s-close", !this.$expanded);
		this.itemsField.domItem.style.display = this.$expanded ? "" : "none";
	},
	onItemEnterLeave: function(onEnter) {
		if (this.itemsField && this.itemsField.menusSlot) {
			this.page.toggleCssOnEnter(this.itemsField.menusSlot, onEnter);
		}
		this.page.toggleCssOnEnter(this.titleField.boxPickers, onEnter);
		this.page.toggleCssOnEnter(this.articleParent.articleParent.menusSlot, onEnter);
		this.page.toggleCssOnEnter(this.articleParent.articleParent.reorderPicker, onEnter);
	},
	_createMenusBox: function() {

	},
	applyActionLinkChange: function(resources) {
		ChildFieldRecord.prototype.applyActionLinkChange.call(this, resources);
		if (!this.menusBox && resources && (resources.$links !== undefined || resources.$actions !== undefined)) {
			this._createMenusBox();
		}
	},
	dispose: function() {
		if (this.blockCardHeader) {
			this.blockCardHeader.syrainout = null;
		}
		this.blockCard = this.titleField = this.itemsField = this.blockCardHeader = null;
		ChildFieldRecord.prototype.dispose.call(this);
	}
});
