// syracuse-ui/lib/site/bookmarks/bookmarkItem/bookmarkItemRecord.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var SingleRecord = require('syracuse-ui/lib/field/array/singleBuilder/singleRecord').SingleRecord;

function BookmarkItemRecord() {}

exports.BookmarkItemRecord = helpers.defineClass(BookmarkItemRecord, SingleRecord, {
	reorderItem: function(targetRecord, isAfter) {
		targetRecord.domItem.parentNode.insertBefore(this.domItem, isAfter ? targetRecord.domItem.nextSibling : targetRecord.domItem);
		this.singleField.variantItem.onReorderItem(this);
	},
	_createMenusBox: function() {
		if (!this.menusSlot) {
			this.menusSlot = document.createElement("div");
			this.page.ensureReorderAndMenusVisibility(this);
			this.page.loadNewItem(this.menusSlot, this.list.defineNewIconMenu("$edit", "s-nav-item-link", true), this);
			this.page.loadNewItem(this.menusSlot, this.list.defineNewIconMenu("$cancel-edit", "s-nav-item-link", true), this);
			this.page.loadNewItem(this.menusSlot, this.list.defineNewIconMenu("$delete", "s-nav-item-link", true), this);
		}
	},
	onMenuClick: function(menuItem) {
		switch (menuItem.$bind) {
			case "$delete":
				if (!this.$isSeparator) {
					menuItem.$confirm = this.localize["nvpConfirmDelete" + (this.singleField.$variantItemKey == "menuItem" ? "Menu" : "Block")];
				}
				break;
			case "$cancel-edit":
				this.switchEditTitle(false);
				return false;
			case "$edit":
				this.switchEditTitle(true);
				return false;
		}
		return true;
	},
	switchEditTitle: function(isEdit) {
		var cell = this.singleField.variantItem.domItem;
		if (isEdit) {
			cell.style.display = "none";
			this.editTitleSlot = document.createElement("div");
			this.editTitleSlot.className = cell.className;
			cell.parentNode.insertBefore(this.editTitleSlot, cell);
			this.editTitleField = this.page.loadNewItem(this.editTitleSlot, {
				$category: "field",
				$isTitleHidden: true,
				$isTopLabelAlignment: false,
				//$contentEditable: true,
				$isEditMode: true,
				$inplace: true,
				$css: "s-nav-admin-menu-edit",
				$field: {
					$isMandatory: true,
					$type: "application/x-string",
					$displayLength: 25
				}
			}, this);
			this.editTitleField.focus();
			this.editTitleField.setDataValue(this.singleField.variantItem.getDatasetValue().title);
		} else {
			cell.style.display = "";
			if (this.editTitleField) {
				this.removeItem(this.editTitleField, true, true);
				this.editTitleField = this.editTitleSlot = null;
			}
		}
		this.applyChange({
			$actions: {
				$edit: {
					$isHidden: isEdit
				},
				"$cancel-edit": {
					$title: this.localize.bookmarksCloseEdit,
					$isHidden: !isEdit
				}
			}
		});
	},
	onNotifyDataChange: function(field, value) {
		if (field == this.editTitleField) {
			var menuItem = this.dataset.$singleField.menuItem;
			menuItem.title = menuItem.$title = value;
			this.applyChange(this.dataset);
			return true;
		}
		return false;
	},
	onItemEnterLeave: function(event) {
		if (this.singleField.variantItem) {
			this.singleField.variantItem.onItemEnterLeave(event, this);
		}
	},
	drawBox: function() {
		this.$isVerticalDirection = true;
		this.domItem = document.createElement("div");
		this.$$item = $(this.domItem);
		this.domItem.className = this.list.$skin + "-item";
		this.domItem.setAttribute("data-s-inout", this.id);
		this.domItem.setAttribute("data-s-record", this.$uuid);

		this.recordValue = document.createElement("div");
		this.recordValue.className = this.list.$skin + "-item-value";
		this.$$value = $(this.domItem.appendChild(this.recordValue));

		if (this.list.$capability && this.list.$capability.reorder) {
			this.applyReorderCapability(true);
		}
		this.singleField = this.page.loadNewItem(this.recordValue, {
			$isTitleRowHidden: true,
			//$isMenusHidden: true,
			$isDetailLinkDisabled: true,
			$bind: "$singleField",
			$isEditMode: this.$isEditMode,
			$inplace: true,
			$isDiagnoseInline: true
		}, this);
		this.list.onAppendRecord(this);
		this.applyRecordCapabilities();
		this.page.ensureReorderAndMenusVisibility(this);
		var menuItem = this.dataset.$singleField.menuItem;
		if (menuItem && !menuItem.$isSeparator) {
			this.applyChange({
				$actions: {
					$edit: {
						$title: this.localize.bookmarksEdit
					}
				}
			});
		}
	},
	applyReorderCapability: function(reorder) {
		if (this.hasReorderCapability = reorder) {
			if (!this.reorderPicker) {
				this.reorderPicker = document.createElement("div");
				this.reorderPicker.className = this.list.$skin + "-item-reorder";
				this.reorderPicker.syraReorderRecordId = this.list.id;
				document.site.ddManager.setDragSpot(this.reorderPicker, true);
				//this.domItem.insertBefore(this.reorderPicker, this.recordValue);
				if (this.list.$isDisabled) {
					document.site.disableItem(this.reorderPicker, this.list.$isDisabled);
				}
			}
		} else {
			if (this.reorderPicker) {
				document.site.removeDomChild(this.reorderPicker);
				delete this.reorderPicker;
			}
		}
	},
	dispose: function() {
		this.editTitleSlot = this.editTitleField = this.singleField = this.$$value = this.domItem = null;
		SingleRecord.prototype.dispose.call(this);
	}
});
