// syracuse-ui/lib/site/bookmarks/bookmarkItem/bookmarkItemList.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var NavigationItemList = require("syracuse-ui/lib/page/navigation/navigationItem/navigationItemList").NavigationItemList;
var BookmarkItemRecord = require('./bookmarkItemRecord').BookmarkItemRecord;

function BookmarkItemList() {}

exports.BookmarkItemList = helpers.defineClass(BookmarkItemList, NavigationItemList, {
	initializeNavigationList: function() {
		this.$isMenuBarAtTop = true;
		this.RecordClass = BookmarkItemRecord;
		this.columnsCount = 1;
		this.emptyMessage = syra_local.bookmarksNoMenuItem;
		this.$item.$isPickerMenuHidden = true;
		this.$item.$isTopLabelAlignment = true;
		this.$item.$skin = "s-bookmarks-items";
		this.$itemRecordMenuSkin = "s-bookmarks-item-link";
	},
	_ensureMenuSlot: function() {
		if (!this.menusSlot) {
			this.menusSlot = document.createElement("div");
			this.menusSlot.className = "s-list-menus-slot";
			var bar = this.ensureTopBar();
			bar.insertBefore(this.menusSlot, bar.firstChild);
		}
		return this.menusSlot;
	},
	onAddMenuItem: function($bind, $menu) {
		return null;
	}
});
