// syracuse-ui/lib/site/bookmarks/bookmarkItem/bookmarkItemList.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var NavigationItemList = require("syracuse-ui/lib/page/navigation/navigationItem/navigationItemList").NavigationItemList;
var BookmarkItemRecord = require('./bookmarkItemRecord').BookmarkItemRecord;

function BookmarkItemList() {}

exports.BookmarkItemList = helpers.defineClass(BookmarkItemList, NavigationItemList, {
	initializeNavigationList: function() {
		this.$isMenuBarAtTop = true;
		this.RecordClass = BookmarkItemRecord;
		this.columnsCount = 1;
		this.emptyMessage = syra_local.bookmarksNoMenuItem;
		this.$item.$isPickerMenuHidden = true;
		this.$item.$isTopLabelAlignment = true;
		this.$item.$skin = "s-nav-admin-items";
		this.$itemRecordMenuSkin = "s-mn-ldp-record-link";
	},
	_ensureMenuSlot: function() {
		if (!this.menusSlot) {
			this.menusSlot = document.createElement("div");
			this.menusSlot.className = "s-list-menus-slot";
			this.topbar.insertBefore(this.menusSlot, this.topbar.firstChild);
		}
		return this.menusSlot;
	},
	onAddMenuItem: function($bind, $menu) {
		return $bind == "addSeparator" && NavigationItemList.prototype.onAddMenuItem.call(this, $bind, $menu);
	},
	onMenuClick: function(menuItem) {
		if (menuItem.$sourceBind == "addSeparator") {
			this.ensureDataSet().push({
				menuItem: {
					$isSeparator: true
				}
			});
			this.setDataBind(this.dataset, this.articleParent.ensureDataSet());
			this.page.notifyDataChange();
			return false;
		}
		return NavigationItemList.prototype.onMenuClick.call(this, menuItem);
	}
});
