// syracuse-tablet/html/js/pages/sdata/pageHome.js
"use strict";

var Base = require('syracuse-tablet/html/js/pages/sdata/pageSdataDashboard').Page;
var BaseSdata = require('syracuse-tablet/html/js/pages/sdata/pageSdata').Page;
var utils = require('syracuse-tablet/html/js/helpers/utils');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var HomeController = require("syracuse-tablet/html/js/controllers/homeController").HomeController;
var native = require('syracuse-tablet/html/js/helpers/native/native');
var wpHelpers = require('syracuse-tablet/html/js/sdata/wpHelpers');
var modules = require('syracuse-tablet/html/js/common/modules');
var metaDataCache = require('syracuse-tablet/html/js/sdata/cache/metaDataCache');
var authHelpers = require('syracuse-tablet/html/js/authoring/authoringHelpers');

var ModalChooseApps = require('syracuse-tablet/html/js/ui/modals/modalChooseApps').Modal;
var ModalChooseDashboardGroup = require('syracuse-tablet/html/js/ui/modals/modalChooseDashboardGroup').Modal;
var ModalConfigTile = require('syracuse-tablet/html/js/ui/modals/modalConfigTile').Modal;
var ModalConfirm = require('syracuse-tablet/html/js/ui/modals/modalConfirm').Modal;
var ModalChooseWelcomeDashboard = require('syracuse-tablet/html/js/ui/modals/modalChooseWelcomeDashboard').Modal;
var ModalChooseRoles = require('syracuse-tablet/html/js/ui/modals/modalChooseRoles').Modal;
var notifications = require('syracuse-tablet/html/js/helpers/notifications');

/**
 *
 */
exports.Page = utils.defineClass(
	function PageHome(pageData, options) {
		Base.call(this, pageData, options);
		this.actionAdapter.showDesignPage = false;
		this.actionAdapter.showPageShare = false;
		this.actionAdapter.showPinPage = false;
		this.actionAdapter.showHome = false;
		this.actionAdapter.showPageConfig = true;
		this.actionAdapter.showMyTemplates = this.controller.showMyTemplates();
		this.actionAdapter.showMyDashboard = true;;
		this.nativeVoiceCommands = native.getModule("voiceCommands");
		notifications.subscribe(this, ["sm.dashboard.description.change", "sm.vignette.pressed"]);
	},
	Base, {
		destroy: function() {
			Base.prototype.destroy.call(this);
		},
		getPageTitle: function() {
			return this.isPersonal() ? locale.text("global.pageTitle.home") : this.pageData.meta.description;
		},
		isRootPage: function() {
			return true;
		},
		isPersonal: function() {
			return this.pageData.isPersonal;
		},
		roles: function() {
			return this.pageData.page.$roles;
		},
		_ensureController: function() {
			this.controller = new HomeController(this.pageData.dataset, this);
			// Will raise notifications or do confirmations
			this.controller.setUIAdapter(this);
		},

		afterRender: function() {
			Base.prototype.afterRender.call(this);
			this.updateVoiceCommands();

		},

		updateVoiceCommands: function() {
			var self = this;
			if (self.nativeVoiceCommands) {
				self.getVoiceCommands().then(function(commands) {
					self.nativeVoiceCommands.updateVoiceCommands(commands);
				});
			}
		},
		_actPageConfig: function() {
			this.controller.toggleConfig(true);
			this._refresh();
		},
		_actMyTemplates: function() {
			var self = this;
			var views = [];
			var currentViews = self.pageData.page.$views || [];
			currentViews.forEach(function(view) {
				if (view.$factoryOwner && view.$factoryOwner === self.controller.factoryId()) {
					views.push(view)
				}
			});
			var opts = {
				showCreateTemplate: true,
				views: views
			}
			var modal = new ModalChooseWelcomeDashboard(opts);
			modal.show()
				.then(function(result) {
					if (result) {
						if (result.selectedView) { // uuid template
							var opts = {
								uuid: result.selectedView
							};
							modules.get("navHelper").gotoUrl("html://home", null, opts);
						} else {
							var article = {
								"$layoutType": "hub",
								"$items": [{
									$layoutType: "hubgroup",
									$title: locale.text("welcome.newgroup")
								}]
							};
							self.pageData.page.$article = article;
							self.pageData.meta.uuid = null;
							self.pageData.meta.description = locale.text("dashboard.welcome.create.template");
							_cleanupUnreferenced(self.pageData);
							self.pageData.isPersonal = false
							self._enabledConfig(true)
							self._refresh(true);
						}
					}
				})
		},
		_actMyDashboard: function() {
			modules.get("navHelper").gotoUrl("html://home");
		},
		_enabledConfig: function(forceShow) {
			this.controller.enabledConfig(forceShow)
			var $$tiles = $(".s-m-tile", this.$$elmt);
			$$tiles.addClass("s-m-tile-edit")
		},
		_disabledConfig: function() {
			this.controller.disabledConfig();
			$(".s-m-tile-edit", this.$$elmt).removeClass("s-m-tile-edit");
		},
		/**
		 * List of voice commands that can be called by user
		 * will be stored in native api and open the called page if the title matches
		 * 
		 * Promise resolving with:
		 * [
		 * {
		 *   title: "My Customers",
		 *   page: {
		 *     $url: ...
		 *     $method: ...
		 *   }
		 * ]
		 */
		getVoiceCommands: function() {
			return $.smResolve(wpHelpers.welcomePageToCommands(this.pageData));
		},
		getPageDescription: function() {
			return this.pageData.meta.description
		},
		_actCloseDashboardConfig: function() {
			if (this.controller.isDirty()) { // restaure
				var opts = {
					uuid: this.pageData.meta.uuid
				};
				modules.get("navHelper").gotoUrl("html://home", null, opts);
			} else {
				this.controller.toggleConfig();
				this._refresh(false)
			}
		},
		_actSavePersonal: function() {
			var self = this;
			var data = wpHelpers.createWelcomeDashboardFromPageData(this.pageData, true);
			wpHelpers.saveWelcomeDashboard(data).then(function(res) {
				self.controller.cleanContext();
				self.controller.setDirty(false);
				modules.get("modal").notify({
					body: locale.text("dashboard.welcome.saved"),
					severityClass: "success"
				});
			})
		},
		_actSaveTemplate: function() {
			var self = this;
			var data = wpHelpers.createWelcomeDashboardFromPageData(this.pageData, false);
			wpHelpers.saveWelcomeDashboard(data)
				.then(function(res) {
					self.pageData.page.$views = res.$views;
					self.pageData.meta.uuid = res.$uuid;
					self.controller.cleanContext();
					self.controller.setDirty(false);
					modules.get("modal").notify({
						body: locale.text("dashboard.welcome.saved"),
						severityClass: "success"
					});
				})
		},
		_actRoleList: function() {
			var self = this;
			_getAvailableRoles(self.roles())
				.then(function(roles) {
					var modal = new ModalChooseRoles(roles);
					return modal.show();
				})
				.then(function(result) {
					if (result && result.change) {
						self.pageData.page.$roles = []
						result.selectedRoles.forEach(function(role) {
							self.pageData.page.$roles.push({
								$uuid: role
							})
						})
						self.controller.rolesChange()
					}
				})
		},
		_actAddGroup: function() {
			// Called when add group in the footer is clicked
			this.getArticle().$items.push({
				$items: [],
				$layoutType: "hubgroup",
				$title: locale.text("welcome.newgroup")
			})
			this._refresh(true)
		},
		_refresh: function(dirty) {
			var self = this;
			self.authUpdateLayout(self.getArticle())
				.then(function() {
					if (self.controller.configActive) { // config already enabled restore it
						self._enabledConfig(true)
					}
					self.controller.setDirty(dirty)
				})
				.fail(function(e) {
					modules.get("modal").error(e);
				});
		},
		renderRootLayout: function() {
			var self = this;
			return self._canAddTile().then(function(canAddTile) {
				self._updateRenderOptions(canAddTile);
				self.rootLayout.set$$container(self.$$contentRoot);
				self.rootLayout.buildHtml();
			})
		},
		_updateRenderOptions: function(canAddTiles) {
			var self = this;
			if (this.controller.isConfigActive()) {
				self.rootLayout.setBuildOptionPerClass("LayoutHubGroup", "showAddTiles", canAddTiles);
				self.rootLayout.setBuildOptionPerClass("LayoutTile", "tileActions", self._defineTileActions());

				var groups = self.rootLayout.children;
				groups.forEach(function(group, idx) {
					var groupActions = [{
						action: "removeGroup",
						className: "remove",
						icon: "fa fa-trash"
					}, {
						action: "moveGroupLeft",
						disabled: idx === 0,
						className: "s-m-move-group" + (idx === 0 ? " disabled" : ""),
						icon: "fa fa-arrow-left"
					}, {
						action: "moveGroupRight",
						disabled: idx === groups.length - 1,
						className: "s-m-move-group",
						icon: "fa fa-arrow-right"
					}];
					group.setBuildOption("groupActions", groupActions);
				});
			} else {
				self.rootLayout.setBuildOptionPerClass("LayoutHubGroup", "showAddTiles", null);
				self.rootLayout.setBuildOptionPerClass("LayoutHubGroup", "groupActions", null);
				self.rootLayout.setBuildOptionPerClass("LayoutTile", "tileActions", null);
			}
		},
		_defineTileActions: function() {
			var tileActions = [];
			if (this._hasMoreThanOneGroup()) {
				tileActions.push({
					action: "changeGroup",
					className: "changegroup",
					icon: "fa fa-sitemap"
				});
			}
			tileActions.push({
				action: "configTile",
				className: "config",
				icon: "fa fa-pencil"
			});
			tileActions.push({
				action: "removeTile",
				className: "remove",
				icon: "fa fa-trash-o"
			});
			return tileActions;
		},
		_hasMoreThanOneGroup: function() {
			return this.getArticle().$items.length > 1;
		},
		authUpdateLayout: function(article) {
			var self = this;
			return BaseSdata.prototype.authUpdateLayout.call(self, article)
		},

		/*
		 * TILE BTN ACTIONS
		 */
		_actChangeGroup: function(tileId) {
			var self = this;
			var grpIdxDest, grpIdx, tileIdx
			var ctrlTile = this.controller.getControl(tileId);
			var groupId = ctrlTile.parent.id;
			var groups = _getOtherGroups(this.rootLayout, groupId);
			var modal = new ModalChooseDashboardGroup(groups);
			modal.show().then(function(groupIdDest) {
				if (groupIdDest) {
					if (self.rootLayout.children.some(function(group, idx) {
							if (group.id === groupIdDest) {
								grpIdxDest = idx;
								return true;
							}
						})) {
						if (self.rootLayout.children.some(function(group, idx) {
								if (group.id === groupId) {
									grpIdx = idx;
									return true;
								}
							})) {
							if (self.rootLayout.children[grpIdx].children.some(function(tile, idx) {
									if (tile.id === tileId) {
										tileIdx = idx;
										return true;
									}
								})) {
								var article = self.getArticle();
								article.$items[grpIdxDest].$items = article.$items[grpIdxDest].$items || []
								article.$items[grpIdxDest].$items.push(article.$items[grpIdx].$items.slice(tileIdx, tileIdx + 1)[0]);
								article.$items[grpIdx].$items.splice(tileIdx, 1)
								self._refresh(true);
							};
						};

					}
				}
			});
		},
		_actConfigTile: function(tileId) {
			var self = this;
			var ti = this._getTileInfo(tileId);
			if (ti) {
				var modal = new ModalConfigTile(ti);
				modal.show().then(function(result) {
					if (result) {
						self._setTileInfo(result);
						self._rebuildPageByControls(true);
					}
				})
			}
		},

		_actRemoveTile: function(tileId) {
			this._removeTile(tileId);
		},

		/*
		 * GROUP BTN ACTIONS
		 */

		_actMoveGroupLeft: function(hubGroupId) {
			this._moveGroup(hubGroupId, -1);
		},
		_actMoveGroupRight: function(hubGroupId) {
			this._moveGroup(hubGroupId, 1);
		},
		_actRemoveGroup: function(hubGroupId) {
			var self = this;
			var modal = new ModalConfirm(locale.text("welcome.dashboard.remove.group"), null, true);
			modal.show().then(function(result) {
				if (result === "yes") {
					self._deleteGroup(hubGroupId);
				}
			});
		},

		_actAddHiddenTile: function(hubGroupId) {
			var self = this;
			_getAvailabeNotUsedApps(this.getArticle(), this.pageData.dataset.json)
				.then(function(notUsedApps) {
					var modal = new ModalChooseApps(notUsedApps);
					return modal.show();
				})
				.then(function(result) {
					// result is a list if selected application objects
					if (result && result.length > 0) {
						return self._addAppsToGroup(result, hubGroupId);
					}
				})
				.then(function(tileAdded) {
					self._rebuildPageByControls(tileAdded);
				});
		},
		_addAppsToGroup: function(apps, hubGroupId) {
			var self = this;
			var hubGroupCtrl = self.controller.getControl(hubGroupId);
			if (!hubGroupCtrl) {
				return false;
			}
			var tileAdded = apps.reduce(function(a, b, index, appList) {
				var vignetteUuid = _addVignetteAndGadget(self.pageData.dataset.json, self.pageData.prototype.json, appList[index])
				b = false;
				if (vignetteUuid) {
					b = _bindVignetteToHubGroupCtrl(hubGroupCtrl, vignetteUuid);
				}
				return a || b

			}, false)
			return tileAdded;
		},
		_moveGroup: function(hubGroupId, dir) {
			var self = this;
			var myIdx;
			if (self.rootLayout.children.some(function(group, idx) {
					if (group.id === hubGroupId) {
						myIdx = idx;
						return true;
					}
				})) {
				var newIdx = myIdx + dir;
				if (newIdx >= 0 && newIdx < self.rootLayout.children.length) {
					var article = self.getArticle();
					var destGrp = article.$items[newIdx];
					article.$items[newIdx] = article.$items[myIdx];
					article.$items[myIdx] = destGrp;
					self._refresh(true);
				}
			}
		},
		// Called by hub group if title canges
		// use to inject title into page main article
		onGroupTitleChanged: function(ctrlHubGroup, uuid, title) {
			this.getArticle().$localization = this.getArticle().$localization || {};
			utils.setLocalization(this.getArticle().$localization, uuid, title);
			this.controller.setDirty(true);
		},
		_removeTile: function(tileId) {
			var self = this;
			var tile = this.controller.getControl(tileId);
			var group = tile.parent;
			var tileIndex = group.getIndexOfChild(tile);
			var hub = group.parent;
			var groupIndex = hub.getIndexOfChild(group);

			var article = self.getArticle();
			article.$items[groupIndex].$items.splice(tileIndex, 1);
			_cleanupUnreferenced(self.pageData);
			self._refresh(true);
		},
		_deleteGroup: function(hubGroupId) {
			var self = this;
			var myIdx;
			if (self.rootLayout.children.some(function(group, idx) {
					if (group.id === hubGroupId) {
						myIdx = idx;
						return true;
					}
				})) {
				var article = self.getArticle();
				article.$items.splice(myIdx, 1);
				_cleanupUnreferenced(self.pageData);
				self._refresh(true);
			}
		},
		_getTileInfo: function(tileId) {
			var tileInfo;
			var ctrlTile = this.controller.getControl(tileId);
			var ctrlVignette = ctrlTile && ctrlTile.getChildByIndex(0);
			if (ctrlVignette && ctrlVignette.$bind) {
				var article = ctrlTile.getArticle();
				var gadgetId = this.pageData.dataset.json[ctrlVignette.$bind];
				var gadget = gadgetId && this.pageData.dataset.json.$mobileGadgets[gadgetId];
				tileInfo = {
					icon: gadget && gadget.icon,
					title: gadget && gadget.$title,
					size: article.$size,
					color: article.$bgColor,
					gadgetId: gadgetId,
					ctrlTile: ctrlTile,
					vignetteId: ctrlVignette.$bind
				};
			}
			return tileInfo;
		},
		_setTileInfo: function(tileInfo) {
			var article = tileInfo.ctrlTile.article = tileInfo.ctrlTile.article || {};
			article.$size = tileInfo.size;
			article.$bgColor = tileInfo.color;
			var gadget = this.pageData.dataset.json.$mobileGadgets[tileInfo.gadgetId];
			if (gadget) {
				gadget.icon = tileInfo.icon;
				gadget.title = tileInfo.title;
			}
			var gadget = this.pageData.prototype.json.$properties[tileInfo.vignetteId].$item;
			if (gadget) {
				gadget.icon = tileInfo.icon;
				gadget.title = tileInfo.title;
			}
		},
		_rebuildPageByControls: function(dirty) {
			var self = this;
			var states = [];
			var newArticle = JSON.parse(authHelpers.rebuildArticleByPageStruct(self, states));
			self.pageData.page.$article = newArticle;
			self._refresh(dirty);
		},
		_canAddTile: function() {
			return _getAvailabeNotUsedApps(this.getArticle(), this.pageData.dataset.json).then(function(appsToAdd) {
				return appsToAdd.length > 0
			})
		},
		getPagePublishOptions: function() {
			return {};
		},
		notifDashboardDescriptionChange: function(newDescription) {
			this.controller.dashboardDescriptionChange(newDescription);
		},
		notifVignettePressed: function(vignetteId) { // long click on tile
			console.log("notifVignettePressed", vignetteId);
			// TODO - Drag and drop
		},
		getPinPageId: function() {
			return this.pageName;
		},
	}
);

/**
 * Return list of application objects that are in the global list of available apps but not yet
 * refered to by the given article
 * 
 */
function _getAvailabeNotUsedApps(article, dataset) {
	var boundApps = _getBoundApps(article, dataset);
	return metaDataCache.getApplicationsList()
		.then(function(appList) {
			var notUsedApps = [];
			appList.forEach(function(app) {
				if (boundApps.indexOf(app.applicationName) < 0) {
					notUsedApps.push(app);
				}
			});
			return wpHelpers.filterApps(notUsedApps);
		});
}

function _getAvailableRoles(currentRoles) {
	var activeRoles = {};
	if (currentRoles) {
		activeRoles = currentRoles.reduce(function(roles, current) {
			roles[current.$uuid] = "active";
			return roles;
		}, {});
	};
	return wpHelpers.getListOfRoles()
		.then(function(roles) {
			roles.forEach(function(role) {
				if (activeRoles[role.$uuid]) {
					role.$selected = activeRoles[role.$uuid]
				}
			});
			return roles;
		})
}
/**
 * Return the list of applicationNames that are bound in the current $article of the page
 * 
 */
function _getBoundApps(article, dataset) {
	var boundApps = [];

	function _checkBind($bind) {
		var gadgetId = dataset[$bind];
		var gadget = gadgetId && dataset.$mobileGadgets[gadgetId];
		if (gadget && gadget.$type === "$application") {
			boundApps.push(gadget.applicationName);
		}
	}
	_getAllBinds(article).forEach(function($bind) {
		_checkBind($bind);
	});
	return boundApps;
}

function _getAllBinds(article) {
	var binds = [];

	function _walk(node) {
		if (node.$bind) {
			binds.push(node.$bind);
		}
		if (node.$items) {
			node.$items.forEach(function($item) {
				_walk($item);
			});
		}
	}
	_walk(article);
	return binds;
}
/**
 * Create new tile article and append it to hubgroupcontrol
 */
function _bindVignetteToHubGroupCtrl(hubGroupCtrl, vignetteUuid) {
	var article = hubGroupCtrl.getArticle();

	var tileArticle = {
		$bgColor: "darkgrey",
		$layoutType: "tile",
		$size: "medium"
	};
	var tile = modules.get("ctrlFactory").createControl(hubGroupCtrl.controller, tileArticle, hubGroupCtrl, hubGroupCtrl.page);
	var vignetteArticle = {
		$bind: vignetteUuid
	};
	var vignette = modules.get("ctrlFactory").createControl(tile.controller, vignetteArticle, tile, tile.page);
	return true;
}

function _addVignetteAndGadget(datasetJson, prototypeJson, app) {
	var gadget = {
		$type: "$application",
		$title: app.title,
		applicationName: app.applicationName,
		icon: app.iconName
	};
	var gadgetId = "$welcome_" + app.applicationName + "_GADGET";
	var vignetteId = "$welcome_" + app.applicationName + "_VIGNETTE";

	datasetJson.$mobileGadgets[gadgetId] = gadget;
	datasetJson[vignetteId] = gadgetId;
	if (prototypeJson) {
		prototypeJson.$properties[vignetteId] = {
			$item: gadget,
			$title: app.title,
			$type: "application/x-vignette-link"
		}
	}
	if (datasetJson.$vignettes) {
		datasetJson.$vignettes.push({
			persistentId: vignetteId,
			gadget: {
				$uuid: gadgetId
			}
		});
	}
	return vignetteId;
}

function _getOtherGroups(rootLayout, groupId) {
	var groups = [];
	rootLayout.children.forEach(function(group) {
		if (group.id !== groupId) {
			groups.push({
				title: group.getArticleText("$title"),
				id: group.id
			});
		}
	});
	return groups;
}

/**
 * Remove vignettes and gadgets not referenced by article any more
 * 
 */
function _cleanupUnreferenced(pageData) {
	var binds = _getAllBinds(pageData.page.$article);
	var boundVignettes = [];
	pageData.dataset.json.$vignettes.forEach(function(vignette) {
		if (binds.indexOf(vignette.persistentId) > -1) {
			boundVignettes.push(vignette);
		} else {
			var vid = vignette.persistentId;
			var gid = vignette.gadget.$uuid;
			delete pageData.dataset.json[vid];
			delete pageData.dataset.json.$mobileGadgets[gid];
		}
	})
	pageData.dataset.json.$vignettes = boundVignettes;
}
