// syracuse-tablet/html/js/pages/sdata/pageSdataQuery.js
"use strict";

var Base = require('syracuse-tablet/html/js/pages/sdata/pageSdata').Page;
var utils = require('syracuse-tablet/html/js/helpers/utils');
var jsutils = require('syracuse-tablet/html/js/helpers/jsutils');
var MultiSelectionController = require("syracuse-tablet/html/js/multiselection/multiSelectionMainController").MultiSelectionMainController;
var DownloadController = require("syracuse-tablet/html/js/multiselection/downloadController").DownloadController;

/**
 * Page handling a query
 * We should add filters and paging
 * 
 */
exports.Page = utils.defineClass(
	function PageSdataQuery(pageData, options) {
		options = options || {};
		options.nativeCapabilities = {
			"liveTile": true,
			"scheduleMessage": true,
			"share": true
		};
		Base.call(this, pageData, options);
		this.actionAdapter.showSortFilter = true; // todo take account authoring;
		this.actionAdapter.showSearch = true; // todo take account authoring;
		this.multiSelectionController = null;
		this.downloadController = null;
	},
	Base, {
		destroy: function() {
			Base.prototype.destroy.call(this);
			if (this.multiSelectionController) {
				this.multiSelectionController.destroy();
				this.multiSelectionController = null;
			}
			if (this.downloadController) {
				this.downloadController.destroy();
				this.downloadController = null;
			}
		},
		_actMultiSelToggle: function() {
			if (!this.attachedControlsMgr) {
				return;
			}
			if (!this.multiSelectionController) {
				this.multiSelectionController = new MultiSelectionController(this);
			}
			if (this.multiSelectionController.multiSelToggleStatus() !== true) {
				this.$$contentElmt.find(".s-m-multi-selected").removeClass("s-m-multi-selected");
			}
		},
		_actSortFilter: function() {
			var array = this.getQueryArray();
			if (!array) return;
			array._openFilterSortPanel();
		},
		multiSelectionIsEnabed: function() {
			return this.multiSelectionController != null && this.multiSelectionController.isEnabled;
		},
		getQueryArray: function() {
			var array = this.controller.getControlsByType("application/x-array");
			if (array.length !== 1) {
				throw new Error("One and only one array is expected in a query facet")
			}
			return array[0];
		},
		isDownloadEnabled: function() {
			return this.isStateLessMode() && this.getAuthoring("$allowDownload") === true;
		},
		_actMultiSelTriggerDownload: function() {
			if (!this.downloadController) {
				this.downloadController = new DownloadController(this);
			}
			this.downloadController.multiSelTriggerDownload()
		},
		_actMultiSelTriggerAction: function(params, event) {
			if (!this.multiSelectionIsEnabed()) {
				return;
			}
			this.multiSelectionController.multiSelTriggerAction(params, event)
		},
		/**
		 * Used by paging to allow the page to reload itself and update the history also
		 */
		fetchNewPageData: function($link) {
			this._oldX3FilterAuth = this.getQueryArray().$filtersGetAuthoring();
			return Base.prototype.fetchNewPageData.call(this, $link);
		},
		authUpdateLayout: function(article) {
			var step = $.smResolve();
			if (this.getQueryArray().$filtersGetAuthoring() == "none" && this._oldX3FilterAuth != "none") {
				// In case $filters authoring as changed from list/tabs to none we've to remove the filter params from the page's url
				var url = jsutils.parseURL(this.pageData.$url);
				if (url.query && url.query.filter) {
					url.query.filter = "";
					step = this.fetchNewPageData({
						$url: jsutils.urlToString(url)
					})
				}
			}
			var self = this;
			return step.then(function() {
				return Base.prototype.authUpdateLayout.call(self, article);
			})
		}
	});
