// syracuse-tablet/html/js/pages/sdata/pageSdataRowDetail.js
"use strict";

var Base = require('syracuse-tablet/html/js/pages/sdata/pageSdataChildPage').Page;
var utils = require('syracuse-tablet/html/js/helpers/utils');
var modules = require('syracuse-tablet/html/js/common/modules');

/**
 * Page handling a query
 * We should add filters and paging
 * 
 */
exports.Page = utils.defineClass(
	function PageSdataRowDetail(pageData, options) {
		Base.call(this, pageData, options);
		if (!this.options.childContext.rowControl) {
			throw new Error("Unexpected null rowControl");
		}
		// create/edit/details
		this._activity = this.options.childContext.activity;
		this.actionAdapter.showAccept = ["create", "edit"].indexOf(this._activity) >= 0;
		// Cancel deactivated for $edit since backup and revert is not fully implemented yet
		this.actionAdapter.showCancel = ["create"].indexOf(this._activity) >= 0;

		// No row detail authoring at the moment
		this.actionAdapter.showDesignPage = false;

		// Controller was create in the ctor of one of the super classes
		// so we can access it here.
		this.datasetBackupJson = $.extend(true, {}, this.controller.dataset.json);
	},
	Base, {
		destroy: function() {
			Base.prototype.destroy.call(this);
			this.options.childContext.rowControl = null;
		},
		getPageTitle: function() {
			return this.controller.dataset.parent.prototype.getValueByPath("$title", true);
		},
		_actAccept: function() {
			if (!this.options.childContext.rowControl) {
				return;
			}
			var self = this;
			setTimeout(function() {
				self.options.childContext.rowControl.notifyRowSaved();
			})
		},
		_actHistoryBack: function() {
			this._cancel();
		},
		// Not finalized yet, just a draft that will be used by cancel on row
		// detail pages to revert a working copy
		_actCancel: function() {
			this._cancel();
		},
		_cancel: function() {
			if (this._activity === "create") {
				this.options.childContext.rowControl.removeRowDetail()
			} else {
				// TODO: How to cancel an edit in WC mode?

				this.controller.applyDatasetBackup(this.datasetBackupJson)
					.fail(function(result) {
						modules.get("modal").error(result);
					});
			}
			return modules.get("appController").App.goBack();
		}
	}
);
