// syracuse-ui/lib/page/bar/barResizerDDAgent.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function BarResizerDDAgent() {}

exports.BarResizerDDAgent = helpers.defineClass(BarResizerDDAgent, null, {
	start: function(dropItem) {
		this.barSplitter = dropItem.barSplitter;
		document.site.ddManager.ddAgent = this;
		document.site.getBoundingClientRect(this.barSplitter.barSlot, this);
		this.isLeftDirection = dropItem.resizeDirection == "left";
		this.resizerRect = this.barSplitter.resizePicker.getBoundingClientRect();
		if (this.barSplitter.$isCollapsed) {
			this.barSplitter.$isCollapsed = false;
			delete this.barSplitter.$barWidth;
			this.barSplitter.ensureState();
		}
	},
	onDragMouseMove: function(target, event) {
		var cursor = "not-allowed";
		if ((event.pageY >= (this.top - 20)) && (event.pageY <= (this.bottom + 20))) {
			var newWidth = this.width + (this.isLeftDirection ? (event.pageX - this.resizerRect.left) : (this.resizerRect.left - event.pageX));
			if (newWidth > 0) {
				this.barSplitter.barSlot.style.width = (this.newWidth = newWidth) + "px";
				cursor = "e-resize";
			}
		}
		document.site.toggleClass(this.barSplitter.openerPicker, "s-resizing", true);
		document.site.body.style.cursor = cursor;
	},
	onDragMouseUp: function(target, event) {
		this.barSplitter.$barWidth = this.barSplitter.barSlot.clientWidth + "px";
		this.barSplitter.barSlot.style.minWidth = this.resizerRect.width + "px";
		this.barSplitter.barSlot.style.width = this.barSplitter.$barWidth;
		if (this.barSplitter.barSlot.clientWidth < this.resizerRect.width) {
			this.barSplitter.barSlot.style.width = this.resizerRect.width + "px";
		}
		document.site.body.style.cursor = "default";
		document.site.toggleClass(this.barSplitter.openerPicker, "s-resizing", false);
		this.barSplitter.savePreferences("$barWidth");
	},
	dispose: function() {
		this.resizerRect = this.barSplitter = null;
	}
});
