// syracuse-ui/lib/page/bar/barSplitter.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var BarResizerDDAgent = require("./barResizerDDAgent").BarResizerDDAgent;

function _appplyPreferences(bar) {
	var defaultState;
	var defaultViewMode = bar.options.$viewMode || (document.site.isTableDevice() ? "float" : "dock");
	if (document.site.$item.$preferences) {
		defaultState = document.site.$item.$preferences[bar.preferenceKey || bar.userPreferenceKey || "none"];
	}
	defaultState = defaultState || {
		$viewMode: defaultViewMode,
		$isCollapsed: false
	};
	if (bar.preferenceKey) {
		document.site.getPreferences(bar.preferenceKey, defaultState, bar);
	} else {
		if (bar.userPreferenceKey) {
			defaultState = bar.page.$userPreferences[bar.userPreferenceKey] || defaultState;
		}
		document.site.applyValues(bar, defaultState);
	}
	if (bar.options.$viewMode || !bar.$viewMode) {
		bar.$viewMode = defaultViewMode;
	}
}

var _mode = {
	"float": {
		draw: function(bar) {
			var direction = bar.options.resizeDirection || "left";
			bar.barSlot.style.display = "inline-block";
			bar.barSlot.style.position = "absolute";
			bar.barSlot.style.top = "0px";
			bar.barSlot.style[direction] = 0;
			if (direction == "right") {
				bar.page.scrollview.style.marginRight = "10px";
			} else {
				bar.page.scrollview.style.marginLeft = "10px";
			}
			(bar.page.vignetteField ? bar.page.vignetteField.body : document.site.body).appendChild(bar.barSlot);
			document.site.setZIndex(bar.barSlot);
		}
	},
	dock: {
		draw: function(bar) {
			var direction = bar.options.resizeDirection || "left";
			bar.barSlot.style.position = "";
			bar.barSlot.style[direction] = "";
			if (direction == "right") {
				bar.page.scrollview.style.marginRight = "";
				bar.page.dataSlot.parentNode.appendChild(bar.barSlot);
			} else {
				bar.page.scrollview.style.marginLeft = "";
				bar.page.dataSlot.parentNode.insertBefore(bar.barSlot, bar.page.dataSlot);
			}
			bar.barSlot.style.display = "";
		}
	}
};

function BarSplitter() {}

exports.BarSplitter = helpers.defineClass(BarSplitter, null, {
	load: function(page) {
		this.minWidth = this.minWidth || 200;
		this.page = page;
		_appplyPreferences(this);
		this._bar = document.createElement("div");
		this._bar.className = this.options.$skins.bar;

		if (!this.options.$isResizerDisabled) {
			this.resizePicker = document.createElement("a");
			this.resizePicker.className = this.options.$skins.resizePicker;
			this.openerPicker = document.createElement("a");
			this.openerPicker.className = this.options.$skins.openerPicker;
			this.openerPicker.setAttribute("data-s-picker", "s-bar-collapse");
			this.page.setArticleId(this.openerPicker);
			this.openerPicker.style.display = "none";
			this.resizePicker.appendChild(this.openerPicker);
			document.site.ddManager.setDragSpot(this.resizePicker, true);
			this._bar.appendChild(this.resizePicker);
			document.site.ddManager.toggleBarResizerObserver(this, true);
		}

		this._bar.appendChild(this.barBody);
		this.barSlot.appendChild(this._bar);
	},
	isDraggable: function(target, event) {
		if (target == this.resizePicker) {
			if (event.target != this.openerPicker) {
				document.site.ddManager.dropableItem = {
					barSplitter: this,
					resizeDirection: this.options.resizeDirection
				};
				document.site.ddManager.start(this, BarResizerDDAgent, this._bar);
				event.syraRetValue = false;
				return;
			}
		}
		return null;
	},
	toggleBar: function(show) {
		this.isHidden = !show;
		if (this.barSlot) {
			this.barSlot.style.display = this.isHidden ? "none" : "";
		}
	},
	ensureState: function(startResize) {
		if (!this.isHidden && !this._designing) {
			if (startResize && this.$isCollapsed) {
				this.$isCollapsed = false;
				delete this.$width;
				this.barSlot.style.width = this.barSlot.getBoundingClientRect().width + "px";
				this.barBody.style.display = "";
			} else {
				if (this.$isCollapsed) {
					this.barSlot.style.width = "";
					this.barBody.style.display = "none";
				} else {
					this.$isCollapsed = false;
					this.barSlot.style.width = Math.max(this.$width || 0, this.minWidth) + "px";
					this.barBody.style.display = "";
				}
				if (this.options && this.options.resizeDirection) {
					_mode[this.$viewMode].draw(this);
				}
			}
			document.site.toggleClass(this.barSlot, "s-close", this.$isCollapsed);
			if (this.resizePicker) {
				document.site.toggleClass(this.openerPicker, "s-close", this.$isCollapsed);
			}
		}
	},
	resetHeight: function() {
		this._bar.style.height = this._bar.style.height = "";
	},
	resizeBar: function() {
		if (this._bar && this.barSlot) {
			var height = this.page.barHeight || this.page.scrollview.getBoundingClientRect().height;
			if (height) {
				this.ensureState();
				this._bar.style.height = height + "px";
				if (this.$viewMode == "float") {
					this.barSlot.style.height = height + "px";
				}
				if (this.resizePicker) {
					this.resizePicker.style.height = height + "px";
					this.openerPicker.style.display = "";
					this.openerPicker.style.top = ((height - this.openerPicker.clientHeight) / 2) + "px";
				}
				this.barBody.style.height = height + "px";
				if (this.designOverlay) {
					var style = this.designOverlay.style;
					var barRect = this._bar.getBoundingClientRect();
					style.top = barRect.top + "px";
					style.left = barRect.left + "px";
					style.height = barRect.height + "px";
					style.width = barRect.width + "px";
				}
				return true;
			}
		}
		return false;
	},
	collapse: function() {
		this.$isCollapsed = !this.$isCollapsed;
		this.savePreferences("$isCollapsed");
		this.ensureState();
		this.page.resizeArticle();
	},
	savePreferences: function(key) {
		if (this.preferenceKey) {
			document.site.setPreferences(this.preferenceKey, this, [key]);
		} else {
			if (this.userPreferenceKey) {
				(this.page.$userPreferences[this.userPreferenceKey] = this.page.$userPreferences[this.userPreferenceKey] || {})[key] = this[key];
				this.page.savePageDesign(true);
			}
		}
	},
	onClickPicker: function(picker) {
		if (picker == this.openerPicker) {
			this.collapse();
			return true;
		}
		return false;
	},
	onPageResize: function() {
		var autoViewMode = (window.innerWidth < 900 || (window.innerHeight > window.innerWidth)) ? "float" : "dock";
		if (this.autoViewMode == undefined) {
			if (this.$viewMode == autoViewMode) {
				this.autoViewMode = autoViewMode;
			}
		}
		if (this.autoViewMode != autoViewMode) {
			this.$viewMode = this.autoViewMode = autoViewMode;
			this.$isCollapsed = this.$viewMode == "float";
			this.ensureState();
		}
		this.resizeBar();
	},
	dispose: function() {
		if (this.resizePicker) {
			document.site.ddManager.toggleBarResizerObserver(this, false);
		}
		if (this.designOverlay) {
			document.site.removeDomChild(this.designOverlay);
			delete this.designOverlay;
		}
		if (this.barSlot && this.$viewMode == "float") {
			document.site.removeDomChild(this.barSlot);
		}
		this.page = this.barSlot = this.barBody = this._bar = this.resizePicker = this.openerPicker = null;
	}

});
