// syracuse-ui/lib/page/bar/barSplitter.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function _appplyPreferences(bar) {
	var defaultState;
	var defaultViewMode = bar.options.$viewMode || (syra_site.isTabletDevice ? "float" : "dock");
	if (syra_site.$item.$preferences) {
		defaultState = syra_site.$item.$preferences[bar.preferenceKey || bar.userPreferenceKey || "none"];
	}
	defaultState = defaultState || {
		$viewMode: defaultViewMode,
		$isCollapsed: false
	};
	if (bar.preferenceKey) {
		syra_site.getPreferences(bar.preferenceKey, defaultState, bar);
	} else {
		if (bar.userPreferenceKey) {
			defaultState = bar.page.userPreferences.getForKey(bar.userPreferenceKey) || defaultState;
		}
		syra_site.applyValues(bar, defaultState);
	}
	if (bar.options.$viewMode || !bar.$viewMode) {
		bar.$viewMode = defaultViewMode;
	}
}

function Moving(sideBar) {
	syra_site.isInoutDisabled = true;
	this.sideBar = sideBar;
	syra_site.dom.getBoundingClientRect(sideBar.barSlot, this);
	this.isLeftDirection = sideBar.options.resizeDirection == "left";
	if (syra_site.isDocumentRTL) {
		this.isLeftDirection = !this.isLeftDirection;
	}
	this.resizerRect = sideBar.resizePicker.getBoundingClientRect();
	if (sideBar.$isCollapsed) {
		sideBar.ensureState(true);
	}
	syra_site.layoutSlot.style.cursor = "e-resize";
	this.bindMouseEvent(true);
}

helpers.defineClass(Moving, null, {
	_move: function(event) {
		var cursor = "not-allowed";
		if ((event.pageY >= (this.top - 20)) && (event.pageY <= (this.bottom + 20))) {
			var newWidth = this.width + (this.isLeftDirection ? (event.pageX - this.resizerRect.left) : (this.resizerRect.left - event.pageX));
			if (newWidth > 0) {
				this.sideBar.setParentSpace(this.newWidth = newWidth);
			}
		}
	},
	bindMouseEvent: function(bind) {
		var self = this;
		if (bind) {
			syra_site.layoutSlot.addEventListener("mousemove", self.on_mousemove = function(event) {
				event.preventDefault();
				if (!self.timeout) {
					self.timeout = setTimeout(function() {
						self.timeout = null;
						self._move(event);
					}, 50);
				}
			}, false);
			syra_site.layoutSlot.addEventListener("mouseup", self.on_mouseup = function(event) {
				self.end();
			}, false);
		} else {
			self.on_mousemove && syra_site.layoutSlot.removeEventListener("mousemove", self.on_mousemove);
			self.on_mouseup && syra_site.layoutSlot.removeEventListener("mouseup", self.on_mouseup);
			self.on_mousemove = self.on_mouseup = null;
		}
	},

	end: function() {
		this.bindMouseEvent(false);
		if (self.timeout) {
			clearTimeout(self.timeout);
			self.timeout = null;
			this._move(event);
		}
		var newWidth = this.sideBar.barSlot.clientWidth;
		if (newWidth <= this.sideBar.closeWidth) {
			this.sideBar.$isCollapsed = true;
			this.sideBar.ensureState();
		} else {
			this.sideBar.$width = this.sideBar.barSlot.clientWidth;
		}
		syra_site.layoutSlot.style.cursor = "default";
		this.sideBar.savePreferences();
		this.sideBar.page.resizeArticle(true);
		this.dispose();
		syra_site.isInoutDisabled = false;
	},

	dispose: function() {
		this.bindMouseEvent(false);
		if (this.sideBar) {
			delete this.sideBar.moving;
		}
		this.sideBar = this.resizerRect = null;
	}
});

function BarSplitter() {}

exports.BarSplitter = helpers.defineClass(BarSplitter, null, {
	load: function(page) {
		var self = this;
		self.articleParent = page;
		if (!self.id) {
			self.id = page.id + "-" + (++page._childItemOffset);
		}
		syra_store.add(self);
		self.closeWidth = 100;
		self.minWidth = self.minWidth || 200;
		self.page = page;
		if (self.isSlotVisible === undefined) {
			self.isSlotVisible = true;
		}
		_appplyPreferences(self);
		self.options.resizeDirection = self.options.resizeDirection || "left";
		self.isVertical = self.options.resizeDirection == "left" || self.options.resizeDirection == "right";
		self.barSlot.className += " s-bar-slot s-bar-slot-" + self.options.resizeDirection;
		if (self.$width != undefined && parseInt(self.$width) <= self.closeWidth) {
			delete self.$width;
		}
		if (self.options.$isResizerDisabled) {
			self.options.isAutoModeDisabled = true;
		}
		if (!self.options.$isResizerDisabled || self.options.$isCollapsable) {
			self.resizePicker = document.createElement("a");
			self.resizePicker.className = self.options.$skin + "-resizer s-bar-resizer-" + self.options.resizeDirection;
			self.openButton = syra_menus.addIconButton(syra_local.bar_collapse, self.options.$skin + "-opener s-bar-opener", "onOpenerClick", "ESC H " + ((self.options.resizeDirection == "left") ? "L" : "R") + ", ESC F11", self.options.resizeDirection + "_arrow");
			self.openButton.syraTool = self.id;
			self.page.setArticleId(self.openButton);
			self.resizePicker.appendChild(self.openButton);
			if (!self.options.$isResizerDisabled) {
				self.resizePicker.addEventListener("mousedown", self.on_mousedown = function(event) {
					event.preventDefault();
					if (event.target != self.openButton) {
						self.moving = new Moving(self);
						event.stopPropagation();
					}
				}, false);
			}
			self.barSlot.appendChild(self.resizePicker);
		}

		self.barSlot.appendChild(self.barBody);
		syra_site.dom.setZIndex(self.barSlot, true);
		self.page.setArticleId(self.barSlot);
		self.ensureState();
	},
	resizeSplitter: function() {
		if (!this.designing) {
			if (!this.options.isAutoModeDisabled) {
				var autoViewMode = syra_site.autoViewMode;
				if (this.autoViewMode == undefined && this.$viewMode == autoViewMode) {
					this.autoViewMode = autoViewMode;
				}
				if (this.autoViewMode != autoViewMode) {
					this.$viewMode = this.autoViewMode = autoViewMode;
					this.$isCollapsed = this.$viewMode == "float";
				}
			}
			this.ensureState();
			this.setParentSpace();
			this.resizeBar();
		}
	},
	setParentSpace: function(newWidth) {
		if (this.isVertical && this.barSlot.parentNode) {
			if (newWidth !== undefined) {
				this.barSlot.style.width = newWidth + "px";
			}
			var direction = this.options.resizeDirection;
			if (this.$isCollapsed || this.$viewMode != "float") {
				if (syra_site.isDocumentRTL) {
					direction = direction == "left" ? "right" : "left";
				}
				if (this.barSlot.parentNode) {
					this.barSlot.parentNode.style["padding" + direction.substr(0, 1).toUpperCase() + direction.substr(1)] = (newWidth || this.barSlot.getBoundingClientRect().width) + "px";
				}
			}
		}
	},
	toggleBar: function(show) {
		this.isHidden = !show;
		if (this.barSlot) {
			this.barSlot.style.display = this.isHidden ? "none" : "";
		}
	},
	ensureState: function(startResize) {
		if (!this.isHidden && this.isVertical) {
			if (startResize && this.$isCollapsed) {
				this.curCollapsedState = this.$isCollapsed = false;
				delete this.$width;
				this.barSlot.style.width = this.barSlot.getBoundingClientRect().width + "px";
				this.barBody.style.display = "";
			} else {
				if (this.curCollapsedState != this.$isCollapsed) {
					if (this.$isCollapsed) {
						this.barBody.style.display = "none";
						this.barSlot.style.width = "";
					} else {
						this.$isCollapsed = false;
						this.barSlot.style.width = Math.max(this.$width || 0, this.minWidth) + "px";
						this.barBody.style.display = "";
					}
					this.curViewMode = this.$viewMode;
					this.curCollapsedState = this.$isCollapsed;
				}
			}
			syra_site.dom.toggleClass(this.barSlot, "s-close", this.$isCollapsed);
			if (this.resizePicker) {
				var icon = this.options.resizeDirection;
				if (this.$isCollapsed) {
					icon = icon == "left" ? "right" : "left";
				}
				syra_menus.updateButtonIcon(this.openButton, this.$isCollapsed ? syra_local.bar_expand : syra_local.bar_collapse, icon + "_arrow");
			}
		}
	},
	resizeBar: function(resize) {
		if (this.barSlot && this.isVertical) {
			var height = this.page.barHeight || this.page.getScrollviewSize().height;
			if (height && (resize || (height != this.lastHeight))) {
				this.lastHeight = height;
				this.barSlot.style.height = height + "px";
				if (this.openButton) {
					this.openButton.style.top = ((height - this.openButton.clientHeight) / 2) + "px";
				}
				this.barBody.style.height = height + "px";
				return true;
			}
		}
		return false;
	},
	collapse: function() {
		this.$isCollapsed = !this.$isCollapsed;
		this.savePreferences();
		this.ensureState();
		this.page.resizeArticle(true);
	},
	clearPreferences: function() {
		var key = this.preferenceKey || this.userPreferenceKey;
		if (key) {
			syra_site.setPreferences(key);
			delete this.$width;
			delete this.$isCollapsed;
			this.ensureState();
			_appplyPreferences(this);
		}
	},
	savePreferences: function(key) {
		if (this.preferenceKey) {
			syra_site.setPreferences(this.preferenceKey, this, ["$width", "$isCollapsed"]);
		} else {
			if (this.userPreferenceKey) {
				var pref = this.page.userPreferences.getForKey(this.userPreferenceKey, {});
				pref.$width = this.$width;
				pref.$isCollapsed = this.$isCollapsed;
				this.page.savePageDesign(true);
			}
		}
	},
	onOpenerClick: function() {
		this.collapse();
	},

	dispose: function() {
		this.on_mousedown && this.resizePicker && this.resizePicker.removeEventListener("mousedown", this.on_mousedown);
		this.moving && this.moving.dispose();
		syra_store.remove(this);
		if (this.barSlot && this.$viewMode == "float") {
			syra_site.dom.removeChild(this.barSlot);
		}
		this.articleParent = this.moving = this.page = this.barSlot = this.barBody = this.resizePicker = this.openButton = null;
	}

});
