// syracuse-ui/lib/page/bar/menuBar.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var BarSplitter = require("syracuse-ui/lib/page/bar/barSplitter").BarSplitter;

function MenuBar() {}

exports.MenuBar = helpers.defineClass(MenuBar, BarSplitter, {
	load: function(page) {
		this.preferenceKey = page.isVignettePage ? "vignette-menubar" : "menuBar";
		this.options = this.options || {};
		this.options.resizeDirection = (page.$pageCategory == "vignette" || page.$pageCategory == "dashboard") ? "left" : "right";
		this.options.$skins = {
			bar: page.$skin + "-menubar",
			openButton: page.$skin + "-menubar-opener",
			resizePicker: page.$skin + "-menubar-resizer"
		};
		page.initializeNewItem(this);
		this.barSlot = document.createElement("div");
		this.barSlot.className = page.$skin + "-menubar-slot";
		this.barSlot.style.display = "none";

		this.barBody = document.createElement("div");
		this.barBody.className = page.$skin + "-menubar-body";

		if (page.$pageCategory == "vignette" || page.$pageCategory == "dashboard") {
			this.page.domItem.insertBefore(this.barSlot, this.page.domItem.firstChild);
		} else {
			this.page.domItem.appendChild(this.barSlot);
		}
		this.isSlotVisible = false;
		BarSplitter.prototype.load.call(this, page);
		this.page.appendPageMenus(this.barBody);
	},
	ensureState: function(startResize) {
		if (this.isSlotVisible) {
			BarSplitter.prototype.ensureState.call(this, startResize);
		}
	},
	onMenuDataFilled: function(menuItem) {
		if (!this.isSlotVisible && !this.isHidden) {
			if (menuItem.boxParent.menuGroupRoot) {
				if (menuItem.boxParent.menuGroupRoot.$item.$isMenuPage) {
					this.isSlotVisible = true;
					if (!this.page.designer) {
						this.toggleBar(true);
						this.ensureState();
					}
				}
			}
		}
	},
	toggleBar: function(show) {
		if (!this.isSlotVisible) {
			show = false;
		}
		if (show && this.page.dialogWrapper) {
			if (this.page.isFusionPage) {
				show = this.$facet != "$lookup";
			} else {
				show = false;
			}
		}
		BarSplitter.prototype.toggleBar.call(this, show);
	}
});
