// @sage/syracuse-react/src/page/entity/detailController.tsx
"use strict";
import * as React from 'react';
import { IPageLoadError } from '../pageLoader';
import { AlertPage } from '../../alert/page/alertPage';
import { IPageProps } from '../page';
import { SyraPage, ISyraPageState } from '../syraPage';
import * as Sdata from '../../dataStore/sdata';
import { SyraMainPage } from '../syraMainPage';
import { syra_form } from '../../syra/utility/syra_form';

export class DetailController extends React.Component<IPageProps, ISyraPageState>  {
    syraPage: SyraPage;
    componentDidMount() {
        this.syraPage = new SyraPage();
        this.syraPage.fetchRepresentation(this.props.sdataUrl, (error) => {
            if (error) {
                this.notifyLoadError(error);
            }
            else {
                this.syraPage.fetchData((diagnoses: Sdata.IDiagnose[]) => {
                    if (diagnoses) {
                        this.notifyLoadError({
                            diagnoses: diagnoses
                        });
                    }
                    else {
                        this.syraPage.loadSyraPage(() => this.forceUpdate());
                    }
                });
            }
        });
    }
    notifyLoadError(error: IPageLoadError) {
        this.setState({
            loadError: error
        });
    }
    onViewMount = (body: HTMLElement) => {
        this.syraPage.slotDidMount(body);
        if (this.props.onMainPageLoaded) {
            this.syraPage.addHistoryLinks();
            syra_form.showLastSavingDiagnoses(this.syraPage.page);
            this.props.onMainPageLoaded(this.syraPage.title, this.syraPage.description, this.syraPage.page);
        }
    }
    componentWillUnmount() {
        this.syraPage && this.syraPage.dispose();
    }
    render() {
        if (this.state && this.state.loadError) {
            return <AlertPage {...this.state.loadError} />;
        }
        else {
            if (this.syraPage && this.syraPage.page) {
                return <SyraMainPage onDidMount={this.onViewMount} onResize={this.syraPage.onResize} />;
            }
        }
        return null;
    }
}
