// syracuse-ui/lib/desktop/site/authoring/palette/blockPalette.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var ItemPalette = require("./itemPalette").ItemPalette;


function BlockPalette(){
}

exports.BlockPalette = helpers.defineClass(BlockPalette, ItemPalette, {
    loadBox: function(){
        var localize = this.getLocalize();
        this.$childCategory = "field";
        this.$prototype = {
            "$": {
                $isHidden: {
                    $type: "application/x-boolean",
                    $title: localize.isHidden
                },
                $title: {
                    $type: "application/x-string",
                    $title: localize.boxTitle
                },
                $isMaximizable: {
                    $type: "application/x-boolean",
                    $title: localize.isMaximizable
                },
                $isBoxCollapsable: {
                    $type: "application/x-boolean",
                    $title: localize.isBoxCollapsable
                },
                $isTitleHidden: {
                    $type: "application/x-boolean",
                    $title: localize.isTitleHidden
                },
                $isSeparatorsVisible: {
                    $type: "application/x-boolean",
                    $title: localize.isSeparatorsVisible
                },
                $fieldsIsTitleHidden: {
                    $type: "application/x-boolean",
                    $title: localize.isLabelHidden
                }
            },
            $isSeparatorsVisible: false,
            $fieldsIsTitleHidden: false
        };
        this._formatFieldWidth(true);
        this._formatFieldLabelAlign(true);
        this._formatFieldTextLabelAlign(true);
        this._formatLayout(true);
        this._formatChildLayout(true, this.$childCategory);
        document.itemFactory.initializeItem(this, {
            $category: "section",
            $bind: "$box",
            $isSeparatorsVisible: true,
            $skin: "s-author-designer-item",
            $layout: {
                $layoutType: "columns",
                $layoutSubType: "60,40",
                $items: [this._formatBoxLayoutStack(), {
                    $layoutType: "stack",
                    $items: [{
                        $category: "section",
                        $bind: "$field",
                        $title: localize.allFields,
                        $layout: {
                            $layoutType: "columns",
                            $layoutSubType: "50,50",
                            $items: [{
                                $layoutType: "stack",
                                $items: [this._formatChildLayout(false, this.$childCategory), this._formatFieldWidth(false)]
                            }, {
                                $layoutType: "stack",
                                $items: [{
                                    $bind: "$fieldsIsTitleHidden",
                                    $skin: "s-author-field",
                                    $isEditMode: true,
                                    $isLeftSpaceHidden: true
                                }, this._formatFieldsLabelPositionBlock()]
                            }]
                        }
                    }]
                }]
            
            }
        }, this.authoringView);
        ItemPalette.prototype.loadBox.call(this);
    }
});
