// syracuse-ui/lib/desktop/site/authoring/palette/pagePalette.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var ItemPalette = require("./itemPalette").ItemPalette;



function PagePalette(){
}

exports.PagePalette = helpers.defineClass(PagePalette, ItemPalette, {
    loadBox: function(){
        var self = this;
        var localize = this.getLocalize();
        self.$childCategory = "section";
        self.$prototype = {
            "$": {
                $isSeparatorsVisible: {
                    $type: "application/x-boolean",
                    $title: localize.isSeparatorsVisible
                },
                $fieldsIsTitleHidden: {
                    $type: "application/x-boolean",
                    $title: localize.isLabelHidden,
                    $isHidden: true
                }
            },
            $isSeparatorsVisible: false
        
        };
        self._formatFieldWidth(true);
        self._formatFieldLabelAlign(true);
        self._formatFieldTextLabelAlign(true);
        self._formatLayout();
        self._formatChildLayout();
        document.itemFactory.initializeItem(self, {
            $category: "section",
            $bind: "$box",
            $isSeparatorsVisible: true,
            $skin: "s-author-designer-item",
            $layout: {
                $layoutType: "side",
                $items: [{
                    $category: "section",
                    $title: localize.appearance,
                    $layout: {
                        $items: [{
                            $bind: "$isSeparatorsVisible",
                            $skin: "s-author-field",
                            $isEditMode: true,
                            $isLeftSpaceHidden: true
                        }]
                    }
                }, this._formatAllFieldsSection()]
            
            }
        }, self.authoringView);
        ItemPalette.prototype.loadBox.call(self);
    }
    
});
