// syracuse-ui/lib/desktop/site/authoring/palette/sectionBlockPalette.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var ItemPalette = require("./itemPalette").ItemPalette;


function SectionBlockPalette(){
}

exports.SectionBlockPalette = helpers.defineClass(SectionBlockPalette, ItemPalette, {
    loadBox: function(){
        this.$childCategory = this.authoringView.designedNode.$authoringLevel == "section" ? "block" : "field";
        var localize = this.getLocalize();
        this.$prototype = {
            "$": {
                $isHidden: {
                    $type: "application/x-boolean",
                    $title: localize.isHidden
                },
                $title: {
                    $type: "application/x-string",
                    $title: localize.boxTitle
                },
                $isMaximizable: {
                    $type: "application/x-boolean",
                    $title: localize.isMaximizable
                },
                $isBoxCollapsable: {
                    $type: "application/x-boolean",
                    $title: localize.isBoxCollapsable
                },
                $isTitleHidden: {
                    $type: "application/x-boolean",
                    $title: localize.isTitleHidden
                },
                $isSeparatorsVisible: {
                    $type: "application/x-boolean",
                    $title: localize.isSeparatorsVisible
                },
                $fieldsIsTitleHidden: {
                    $type: "application/x-boolean",
                    $title: localize.isLabelHidden,
                    $isHidden: this.$childCategory == "block"
                }
            },
            $isSeparatorsVisible: false
        };
        this._formatFieldWidth(true);
        this._formatFieldLabelAlign(true);
        this._formatFieldTextLabelAlign(true);
        this._formatLayout();
        this._formatChildLayout();
        document.itemFactory.initializeItem(this, this._formatSectionBlock(), this.authoringView);
        ItemPalette.prototype.loadBox.call(this);
    }
});
