// syracuse-ui/lib/authoring/patterns/patternFactory.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var HeaderTabsPattern = require("./headerTabsPattern").HeaderTabsPattern;


exports.build = function($category, $prototype, $source){
    $source = helpers.object.clone($source, true);
    switch ($category) {
        case "pattern-headerTabs":
            var pattern = new HeaderTabsPattern();
            return pattern.build($prototype, $source);
        default:
            var builder = exports.patterns[$category];
            if (builder) {
                var $rootItems = document.site.authorPage.extractItemSettings($source.$layout);
                return {
                    $layout: builder.build($source, $rootItems)
                };
            }
            break;
    }
    return null;
};

exports.patterns = {
    "pattern-1": {
		isBoxOnly : true,
        build: function($sourceArticle, $rootItems){
            var $layout = {
                $layoutType: "stack",
                $items: []
            };
            if ($rootItems.length > 0) {
                $layout.$items.push($rootItems[0]);
            }
            var tabs = {
                $layoutType: "tabs",
                $items: []
            }
            $layout.$items.push(tabs);
            if ($rootItems.length > 1) {
                for (var ii = 1, jj = Math.min($rootItems.length - 1, 3); ii < jj; ii++) {
                    if (ii < $rootItems.length) {
                        tabs.$items.push($rootItems[ii]);
                    }
                    else {
                        tabs.$items.push({
                            $category: "section"
                        });
                    }
                }
            }
            if ($rootItems.length >= 3) {
                $layout.$items.push($rootItems[$rootItems.length - 1]);
            }
            else {
                $layout.$items.push({
                    $category: "section"
                });
            }
            return $layout;
        }
    },
    "pattern-2": {
        build: function($sourceArticle, $rootItems){
            var $layout = {
                $layoutType: "stack",
                $items: []
            };
            var $row, cellIndex = 0;
            var count = $rootItems.length;
            for (var ii = 0, jj = Math.max($rootItems.length - 1, 10); ii < jj; ii++) {
                if (cellIndex == 0) {
                    $layout.$items.push($row = {
                        $layoutType: "row",
                        $layoutSubType: "33,33,33",
                        $items: []
                    });
                }
                if (ii < count) {
                    $row.$items.push($rootItems[ii]);
                }
                if ((++cellIndex) == 3) {
                    cellIndex = 0;
                }
            }
            return $layout;
        }
    },
    "pattern-3": {
        build: function($sourceArticle, $rootItems){
            var $layout = {
                $layoutType: "stack",
                $items: []
            };
            var $row, cellIndex = 0;
            var count = $rootItems.length;
            for (var ii = 0, jj = Math.max($rootItems.length - 1, 16); ii < jj; ii++) {
                if (cellIndex == 0) {
                    $layout.$items.push($row = {
                        $layoutType: "row",
                        $layoutSubType: "25,25,25,25",
                        $items: []
                    });
                }
                if (ii < count) {
                    $row.$items.push($rootItems[ii]);
                }
                if ((++cellIndex) == 4) {
                    cellIndex = 0;
                }
            }
            return $layout;
        }
    },
    "pattern-4": {
        build: function($sourceArticle, $rootItems){
            var $layout = {
                $layoutType: "row",
                $layoutSubType: "38,62",
                $items: []
            };
            var $row, cellIndex = 0;
            var count = $rootItems.length;
            if (count == 0) {
                $layout.$items.push({
                    $category: "section"
                });
            }
            else {
                $layout.$items.push($rootItems[0]);
            }
            var col2 = {
                $layoutType: "stack",
                $items: []
            };
            $layout.$items.push(col2);
            for (var ii = 1, jj = Math.max($rootItems.length - 1, 16); ii < jj; ii++) {
                if (cellIndex == 0) {
                    col2.$items.push($row = {
                        $layoutType: "row",
                        $layoutSubType: "33,33,33",
                        $items: []
                    });
                }
                if (ii < count) {
                    $row.$items.push($rootItems[ii]);
                }
                if ((++cellIndex) == 3) {
                    cellIndex = 0;
                }
            }
            return $layout;
        }
    }
};
