// syracuse-ui/lib/fusion/breadcrumb.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function Breadcrumb(){
}

exports.Breadcrumb = helpers.defineClass(Breadcrumb, null, {
    load: function(gateway){
        var self = this;
        self.gateway = gateway;
        (self._$$opener = $("#s-fusion-sessions-opener")).bind("click", function(){
            if (!self._popup) {
                self._fillPopup();
                self._popup = document.site.openDialog({
                    $dialogMode: "popup",
                    $$dialog: self._$$panel,
                    content: self.gateway, //a tester
                    position: {
                        my: "right top",
                        at: "right bottom",
                        of: self._$$opener
                    },
                    autocCloseBoundary: "#s-fusion-sessions-opener",
                    onClose: function(){
                        self._$$panel.empty();
                        delete self._popup;
                    }
                });
            }
            else {
                self._closePopup();
            }
        });
        self._$$panel = $("<div id='s-fusion-sessions-pn'/>");
        self._$$panel.delegate(".s-fusion-sheets-opener-children", "click", function(){
            var $$opener = $(this);
            var open = !$$opener.hasClass("s-open");
            $$opener.siblings(".s-fusion-sheets")[0].style.display = open ? "" : "none";
            $$opener.toggleClass("s-open", open);
            return false;
        }).delegate(".s-fusion-sheet-link,.s-fusion-book-link", "click", function(){
            var $$link = $(this);
            var book = self._findBook($$link), selectedSheet;
            if (!$$link.hasClass("s-fusion-book-link")) {
                selectedSheet = book.sheets[$$link.attr("data-s-sheet")];
            }
            self.gateway.activateBook(book, selectedSheet);
            setTimeout(function(){
                self._closePopup();
            }, 100);
            return false;
        });
    },
    _fillPopup: function(){
        var self = this;
        var bookUL = document.createElement("ul");
        bookUL.className = "s-fusion-books";
        self._$$panel.empty()[0].appendChild(bookUL);
        
        self.gateway._books.forEach(function(book){
            var bookDiv = document.createElement("div");
            bookDiv.className = "s-fusion-book";
            var sheetsUL = document.createElement("ul");
            sheetsUL.className = "s-fusion-sheets";
            sheetsUL.style.display = "none";
			var sheetOpener;
            book.sheets.forEach(function(sheet, index){
                var title = sheet.getTitle();
                var sheetLink = document.createElement("a");
                sheetLink.className = "s-fusion-book-link";
                sheetLink.setAttribute("data-s-sheet", index);
                if (index == 0) {
                    sheetOpener = document.createElement("a");
                    sheetOpener.className = "s-fusion-sheets-opener";
                    bookDiv.appendChild(sheetOpener);
                    bookDiv.appendChild(sheetLink);
                    if (book.sheets.length > 1) {
                        title += " (" + (book.sheets.length - 1) + ")";
                        sheetOpener.className += " s-fusion-sheets-opener-children";
                    }
                }
                //s-fusion-sheet-link-selected
                else {
                    var li = document.createElement("li");
                    li.className = "s-fusion-sheets-li";
                    var div = document.createElement("div");
                    div.className = "s-fusion-sheet";
                    sheetsUL.appendChild(li).appendChild(div).appendChild(sheetLink);
                }
                if (sheet == book.selectedSheet) {
                    sheetLink.className += " s-fusion-book-link-selected";
                    if (book == self.gateway.activatedBook) {
                        sheetLink.className += " s-fusion-sheet-activated";
                        sheetsUL.style.display = "";
                        sheetOpener.className += " s-open";
                    }
                }
                sheetLink.textContent = title;
            });
			bookDiv.appendChild(sheetsUL);
            var li = document.createElement("li");
            li.className = "s-fusion-books-li";
            li.setAttribute("data-s-book", book.id);
            bookUL.appendChild(li).appendChild(bookDiv);
        });
    },
    
    _findBook: function($$node){
        var id = ($$node.is("[data-s-book]") ? $$node : $$node.closest("[data-s-book]")).attr("data-s-book");
        return this.gateway.findBookById(id).book;
    },
    _closePopup: function(){
        if (this._popup) {
            this._popup.close();
            delete this._popup;
        }
    },
    onBooksChange: function(books){
        var opener = this._$$opener[0];
        opener.textContent = books.length;
        opener.style.visibility = books.length > 0 ? "visible" : "hidden";
    },
    dispose: function(){
        if (this._$$panel) {
            this._$$panel.undelegate();
        }
        this._closePopup();
        if (this._$$opener) {
            this._$$opener.unbind();
            delete this._$$opener;
        }
        delete this._$$panel;
        delete this.gateway;
    }
});
