// syracuse-ui/lib/fusion/grid.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var List = require("syracuse-ui/lib/desktop/field/collection/list").List;
var GridBuilder = require('./gridBuilder').Builder;

function Grid(){
}

exports.Grid = helpers.defineClass(Grid, List, {
    createBuilder: function(){
        console.log("Grid initialize");
        this.builder = new (this.BuilderClass || GridBuilder)();
        this.builder.list = this;
        this.builder.initialize();
        
        this.$isEditMode = false; // TODO : lv
        this.builder.lineCount(this.$field.$lineCount);
        this.builder.$fieldnl = {
            "$constraints": {
                "$maxLength": 5
            },
            "$title": "",
            "$capability": {
                "frozen": true
            },
            "$isHidden": ((this.$item.$X3) ? this.$item.$X3.$hideLineNumber : false)
        };
        this.$itemnl = {
            "$bind": this.builder.getGridXid() + "_NL"
        };
        this.$fusionPageMeta = this.boxParent.getArticle().$fusionPageMeta;
    },
    drawBox: function(){
        List.prototype.drawBox.call(this);
        this.$$item.addClass("s-field-grid-fusion");
        this.$$dataValue.addClass("s-field-value-grid-fusion");
        this.$$valueSlot.addClass("s-field-value-slot-grid-fusion");
        this.$$core.addClass("s-field-core-grid-fusion");
    },
    fetch: function(options, $location){
        //TODO
        if (options.$itemsPerPage) {
            this.builder.changeLineCount(options.$itemsPerPage);
        }
    },
    setDataBind: function($resources, record, metaData){
        if ($resources) {
            this.dataLoaded = false
            if (record.$isDelta) {
                this.builder.resetSort();
            }
            List.prototype.setDataBind.call(this, $resources, record, null);
            this.builder.resizeRowsContainer();
            if (record.$isDelta) {
                this.builder.reSynch(record);
            }
            this.builder.endBinding(record.$isDelta);
        }
        if (metaData) {
            this.builder.setGuiMods(metaData);
            this.builder.applyGuiMod();
            this.builder.endGuiMods();
        }
    },
    _setCallBack: function(cb){
        this.callBack = cb;
    },
    _getValue: function(record, $bind){
        return record.dataset[$bind];
    },
    setFocus: function(xid, nl, edit, opt){
        this.builder.rows[(nl - 1)].setFocus(xid, edit, opt);
    },
    getValue: function(xid, nl){
        this.builder.rows[(nl - 1)].getValue(xid);
    }
});
