// syracuse-ui/lib/fusion/workBook.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var box = require('syracuse-ui/lib/common/article/box').Box;
var fieldFactory = require('syracuse-ui/lib/fusion/field/fieldFactory');
var util = require('syracuse-ui/lib/fusion/core/client/sap/util');
var requestActions = require('syracuse-ui/lib/fusion/core/client/sap/srvactions');
var sapUtil = util.Fusion;
var syraUtil = util.Syra;
function WorkBook(){
    this.sheets = [];
}

exports.WorkBook = helpers.defineClass(WorkBook, null, {
    releaseMainPage: function(options){
        this.$$slot.empty().append(this.sheets[0].$$container.children());
        this.fusionGateway.closeMenus();
    },
    createField: function($field, $item, boxParent, $class){
        return fieldFactory.create($field, $item, boxParent, $class);
    },
    setDataBind: function(field, $resources, record, metaData){
        if (metaData) {
            if (metaData.tit) {
                field.setTitle(metaData.tit);
            }
            if (metaData.dch && field.removeListItems) {
                field.removeListItems(metaData.dch);
            }
            if (metaData.stt) {
                if (field.setState) {
                    field.setState(sapUtil.getState(metaData.stt));
                }
                // TODO : deal with $isReadOnly
            }
            if (metaData.sty) {
                // TODO
            }
        }
        return true;
    },
    unselectSheet: function(){
        if (this.selectedSheet && this.selectedSheet.$$item) {
            if (!this.selectedSheet.$$fusionMemHost) {
                this.selectedSheet.$$fusionMemHost = $("<div/>");
            }
            this.selectedSheet.$$fusionMemHost.append(this.selectedSheet.$$item);
            this.selectedSheet.selectSheet(false);
            delete this.selectedSheet;
        }
    },
    selectSheet: function(sheet){
        var sheetMeta;
        if (this.selectedSheet && this.selectedSheet != sheet) {
            this.unselectSheet();
        }
        if (document.site.$$body) {
            document.site.$$body[0].style.display = "none";
            (this.selectedSheet = sheet).$$item.appendTo(sheet.$$container = document.site.$$body);
            document.site.$$body[0].style.display = "";
            sheetMeta = syraUtil.getFusionPageMeta(sheet);
            sheet.toggleSheetOverlay(!sheetMeta.controller.isSapActiveWindow(sheetMeta.winModel.getWinId()));
            this.selectedSheet.selectSheet(true);
            if (this.fusionGateway.activatedBook && this.fusionGateway.activatedBook != this) {
                this.fusionGateway.activatedBook.onActivate(false);
                this.fusionGateway.activatedBook = this;
                this.onActivate(true);
            }
        }
    },
    _loadDialog: function($itemPage, $facet, opts){
        var options = opts || {}, self = this, dialog;
        // Set dialog options. TODO : improve options settings according $facet...
        options.$itemPage = $itemPage;
        options.boxParent = this;
        options.onClose = function(validated, dispose) {
            if(!dispose) {
                // Close must be confirm by server. So, we return "false"
                self.fusionSite.controller.trigger("wdgt.win.close", {"target": dialog, "type": "close"});
                // Warning : normaly, server will force window close, and then, client controller "close stub" may call "dispose" method of "Dialog" class and, finally, all should be released safety!
                return false;
            }
            else {
                return true;
            }
        };
        // Update sheets manager
        if (this.sheets.length > 0) {
            this.unselectSheet();
        }
        // Open dialog and return is content (i.e page)
        $itemPage.boxParent = null;
        dialog = document.site.openDialog(options);
        (this._dialogs = this._dialogs || {})[dialog.id] = dialog;
        return dialog._content;
    },
    loadSheet: function($itemPage, $facet, opts){
        this.fusionGateway.closeMenus();
        if(syraUtil.isDialogFacet($facet)) {
            this.selectedSheet = this._loadDialog($itemPage, $facet, opts);
        }
        else {
            if (this.sheets.length == 0) {
                this.selectedSheet = document.site.onMainPageChange($itemPage);
            }
            else {
                this.unselectSheet();
                document.site.$$body[0].style.display = "none";
                $itemPage.boxParent = null;
                $itemPage.$$container = document.site.$$body;
                this.selectedSheet = document.itemFactory.loadPage($itemPage);
                document.site.$$body[0].style.display = "";
            }
        }
        this.selectedSheet.selectSheet(true);
        this.fusionGateway.activatedBook = this;
        this.sheets.push(this.selectedSheet);
        return this.selectedSheet;
    },
    closeSheet: function(event, fusionDispose){
        var sheet = event.data.sheet, found;
        if (sheet) {
            for (var ii = this.sheets.length - 1; ii >= 0; ii--) {
                if (this.sheets[ii] == sheet) {
                    if (sheet == this.selectedSheet) {
                        this.unselectSheet(sheet);
                    }
                    this.sheets.splice(ii, 1);
                    this._disposeSheet(sheet);
                    found = ii;
                    break;
                }
            }
            if (found != undefined && (--found) >= 0) {
                this.selectSheet(this.sheets[found]);
            }
        }
    },
    onActivate: function(satus){
        this.fusionSite.onActivate(satus);
    },
    onReleaseBook: function(event){
        // Called by fusionGatway, on "_releaseBook()"
    },
    _disposeSheet: function(sheet){
        var dialog = null;
        if (sheet && !sheet.disposed) {
            if (sheet.$$fusionMemHost) {
                sheet.$$fusionMemHost.remove();
            }
            delete sheet.externalAdapter;
            if(this._dialogs) {
                 dialog = _.find(this._dialogs, function(dialog) {return dialog._content == sheet;});
            }
            if(dialog) {
                delete this._dialogs[dialog.id];
                document.controller.disposeObject(dialog);                
            }
            else {
                document.controller.disposeObject(sheet);
            }
        }
    },
    onMenuItemClick: function(options){
        var doEvt = true;
        if (this.fusionSite.controller) {
            doEvt = this.fusionSite.controller.trigger("wdgt.link.click", {
                "target": options.menuItem,
                "type": "click"
            });
        }
        if (doEvt) {
            options.doEvent();
        }
    },
    onBoxToggle: function(options){
        var doEvt = true, targetType = null, boxMeta = null, xid = null;
        if (this.fusionSite.controller && (boxMeta = syraUtil.getMetaFromObject(options.box))) {
            targetType = (xid = sapUtil.isFolderTab(boxMeta[sapUtil.metaNameMap.xid])) ? "tabs" : null;
            targetType = xid ? targetType : ((xid = sapUtil.isListTab(boxMeta[sapUtil.metaNameMap.xid])) ? "list" : null);
            if (targetType && !options.onDrawParent && !this.fusionSite.controller.isFixtureMode()) {
                doEvt = this.fusionSite.controller.trigger("wdgt." + targetType + ".toggle", {
                    "target": options.box,
                    "type": "toggle",
                    "data": options
                }, xid);
            }
            if (doEvt) {
                options.doEvent();
                if (targetType) {
                    this.fusionSite.controller.trigger("wdgt." + targetType + ".toggled", {
                        "target": options.box,
                        "type": "toggled",
                        "data": options
                    }, xid);
                }
            }
        }
    },
    onBoxClick: function(options){
        var doEvt = true, targetType = null, boxMeta = null, xid = null;
        if (this.fusionSite.controller && (boxMeta = syraUtil.getMetaFromObject(options.box))) {
            targetType = (xid = sapUtil.isFolderTab(boxMeta[sapUtil.metaNameMap.xid])) ? "tabs" : null;
            targetType = xid ? targetType : ((xid = sapUtil.isListTab(boxMeta[sapUtil.metaNameMap.xid])) ? "list" : null);
            if (targetType && !this.fusionSite.controller.isFixtureMode()) {
                doEvt = this.fusionSite.controller.trigger("wdgt." + targetType + ".click", {
                    "target": options.box,
                    "type": "click",
                    "data": options
                }, xid);
            }
            if (doEvt && options.doEvent) {
                options.doEvent();
            }
            else 
                if (doEvt === false) {
                    options.event.stopImmediatePropagation();
                }
        }
    },
    onFieldEvent: function(options){
        var doEvt = true;
        console.log("**onFieldEvent : [" + options.event.type + "] on field [" + (syraUtil.getMetaFromObject(options.field)).$bind + "]");
        switch (options.event.type) {
            case "keydown":
            case "click":
            case "change":
               doEvt = this.fusionSite.controller.trigger("wdgt.field." + options.event.type, {
                            "target": options.field, "type": options.event.type, "data": {"nativeEvt": options.event}
                    });
                if (!doEvt) {
                    options.event.stopImmediatePropagation();
                }
                break;
            case "keyup":
                break;
            case "focusin":
                break;
            case "focusout":
                break;
        }
        if (doEvt) {
            options.doEvent();
        }
        else 
            if (doEvt === false) {
                options.event.preventDefault();
                options.event.stopPropagation();
                options.event.stopImmediatePropagation();
            }
        return doEvt;
    },
    onSelectRecordEvent: function(options){
        var list = options ? options.field : null, listMeta = null;
        if (list && this.fusionSite.controller && (listMeta = syraUtil.getMetaFromObject(options.field))) {
            // Note :  options{field, list, uuidTarget, isSelected,event, doEvent()
            if (isNaN(options.uuidTarget)) 
                return;
            this.fusionSite.controller.trigger("wdgt.list.selline", {
                "target": list,
                "type": "selline",
                "data": {
                    line: parseInt(options.uuidTarget, 10) + 1,
                    xid: listMeta.$bind,
                    doEvent: options.doEvent
                }
            });
        }
    },
    onFieldClickPicker: function(options){
        var doEvt = this.fusionSite.controller.trigger("wdgt.field.picker", {"target": options.field, "type": "picker", "data": options});
        if (doEvt) {
            options.doEvent();
        }       
    },
    onFieldNotifyChange: function(options){
        var doEvt = false;
        //TODO à revoir peut sans doute mieux faire!
        if(options && options.doEvent && options.field && options.field.$item && options.field.$item.$bind==="$itemsPerPage"){
            doEvt=true;
        }
        if(doEvt) {
         options.doEvent();
         }
    },
    release: function(){
        // Disconnect fusion Site
    },
    dispose: function(){
        box.prototype._closeChildrenDialogs.call(this, true);
        var self = this;
        if (self.fusionSite) {
            self.fusionSite.dispose();
            delete self.fusionSite;
        }
        if (self.sheets) {
            self.sheets.forEach(function(sheet){
                self._disposeSheet(sheet);
            });
            delete self.sheets;
        }
        delete self.selectedSheet;
        delete self.fusionGateway;
    }
});
