// @sage/syracuse-lib/src/sdata/asyncTrack.ts
"use strict";
/* @flow */
/*:: 

import { SDataSeverity, SDataLink } from "./types";

type SDataTracking = {
	phase: string;
	phaseDetail: string;
	progress: number;
}

*/
import { globals } from 'streamline-runtime';

exports.$exported = true;

export function asyncTrack(_, phase, detail, progress) {
	if (globals.context.sdataContext && globals.context.sdataContext.tracker) {
		var tt = globals.context.sdataContext.tracker;
		tt.phase = phase;
		tt.phaseDetail = detail;
		tt.progress = progress;
	}
};

export function addDiagnose(_, severity, message) {
	if (globals.context.sdataContext && globals.context.sdataContext.tracker) {
		var tt = globals.context.sdataContext.tracker;
		tt.addDiagnose(severity, message);
	}
};

export function abortRequested(_) {
	if (globals.context.sdataContext && globals.context.sdataContext.tracker) {
		return globals.context.sdataContext.tracker.abortRequested;
	} else {
		return false;
	}
};

export function addLink(_, name, link) {
	if (globals.context.sdataContext && globals.context.sdataContext.tracker) {
		var t = globals.context.sdataContext.tracker;
		t.$links = t.$links || {};
		t.$links[name] = link;
	}
};
