// syracuse-ui/lib/authoring/page/tree/trees.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var InsertTree = require("syracuse-ui/lib/authoring/page/tree/insertTree").InsertTree;
var ContentTree = require("syracuse-ui/lib/authoring/page/tree/contentTree").ContentTree;

function Trees() {}

exports.Trees = helpers.defineClass(Trees, null, {
	load: function(designer) {
		this.designer = designer;
		this.id = designer.id + "-" + (++designer._childItemOffset);
		syra_map[this.id] = this;
		this.designedArticle = this.designer.designedArticle;
		this.treebarTabs = document.createElement("div");
		this.treebarTabs.className = "s-aw-tlbr-tabs-nav";
		this.treebarBody.appendChild(this.treebarTabs);
		this._selectedTab = null;
		this.designer.designedArticle.garbage.loadFreeField();
		(this.contentTree = new ContentTree()).load(this, this.appendTab(syra_local.aw_viewTypeContent));
		(this.insertTree = new InsertTree()).load(this, this.appendTab(syra_local.aw_insert));
		this._selectTab(this.contentTree.tabItem.tab, true);
	},
	_selectTab: function(tab, select) {
		if (select && this._selectedTab && this._selectedTab.tab != tab) {
			this._selectTab(this._selectedTab.tab, false);
		}
		var tabItem = this.contentTree.tabItem.tab == tab ? this.contentTree.tabItem : this.insertTree.tabItem;
		this._selectedTab = (select) ? tabItem : null;
		tabItem.tab.className = "new-s-aw-tlbr-tab" + (select ? " s-tab-open" : "");
		tabItem.body.style.display = select ? "" : "none";
	},
	onTabTreesClick: function(event, btn) {
		this._selectTab(btn, true);
	},
	appendTab: function(title) {
		var tab = syra_menus.addTextButton(title, "new-s-aw-tlbr-tab", "onTabTreesClick");
		tab.syraTool = this.id;
		var body = document.createElement("div");
		body.style.display = "none";
		body.className = "s-aw-content-tree";
		var tabItem = {
			tab: this.treebarTabs.appendChild(tab),
			body: this.treebarBody.appendChild(body)
		};
		return tabItem;
	},
	fill: function() {
		if (this.designedArticle) {
			this.contentTree.fill();
			this.insertTree.fill();
			if (this._selectedTab && this._selectedTab.tab) {
				this._selectTab(this._selectedTab.tab, true);
			}
		}
	},
	selectNode: function(item, select) {
		this.contentTree.selectNode(item, select);
		this.insertTree.selectNode(item, select);
	},
	findNode: function(event) {
		var tree = this._findTree(event);
		return tree ? tree.findNode(event) : null;
	},
	_findTree: function(event) {
		var target = event.target;
		while (target != null && !target.syraAwTree) {
			target = target.parentNode;
		}
		return target.syraAwTree ? this[target.syraAwTree] : null;
	},
	onResizeBar: function(height) {
		this.treebarBody.style.height = height + "px";
		var bodyHeight = Math.max(height - this.treebarTabs.getBoundingClientRect().height, 20) + "px";
		this.contentTree.tabItem.body.style.height = this.insertTree.tabItem.body.style.height = bodyHeight;
	},
	dispose: function() {
		delete syra_map[this.id];
		if (this.contentTree) {
			this.contentTree.dispose();
		}
		if (this.insertTree) {
			this.insertTree.dispose();
		}
		this.contentTree = this.insertTree = this.designer = this.treebarBody = null;
	}
});
