// hdb/lib-cov/protocol/request/Part.js
/* automatically generated by JSCoverage - do not edit */
if (typeof _$jscoverage === 'undefined') _$jscoverage = {};
if (! _$jscoverage['protocol/request/Part.js']) {
  _$jscoverage['protocol/request/Part.js'] = [];
  _$jscoverage['protocol/request/Part.js'][14] = 0;
  _$jscoverage['protocol/request/Part.js'][16] = 0;
  _$jscoverage['protocol/request/Part.js'][17] = 0;
  _$jscoverage['protocol/request/Part.js'][19] = 0;
  _$jscoverage['protocol/request/Part.js'][21] = 0;
  _$jscoverage['protocol/request/Part.js'][23] = 0;
  _$jscoverage['protocol/request/Part.js'][24] = 0;
  _$jscoverage['protocol/request/Part.js'][25] = 0;
  _$jscoverage['protocol/request/Part.js'][26] = 0;
  _$jscoverage['protocol/request/Part.js'][27] = 0;
  _$jscoverage['protocol/request/Part.js'][31] = 0;
  _$jscoverage['protocol/request/Part.js'][32] = 0;
  _$jscoverage['protocol/request/Part.js'][33] = 0;
  _$jscoverage['protocol/request/Part.js'][35] = 0;
  _$jscoverage['protocol/request/Part.js'][37] = 0;
  _$jscoverage['protocol/request/Part.js'][39] = 0;
  _$jscoverage['protocol/request/Part.js'][41] = 0;
  _$jscoverage['protocol/request/Part.js'][43] = 0;
  _$jscoverage['protocol/request/Part.js'][45] = 0;
  _$jscoverage['protocol/request/Part.js'][46] = 0;
  _$jscoverage['protocol/request/Part.js'][47] = 0;
  _$jscoverage['protocol/request/Part.js'][48] = 0;
  _$jscoverage['protocol/request/Part.js'][50] = 0;
}
_$jscoverage['protocol/request/Part.js'][14]++;
"use strict";
_$jscoverage['protocol/request/Part.js'][16]++;
var util = require("../../util");
_$jscoverage['protocol/request/Part.js'][17]++;
var PartKind = require("../common/PartKind");
_$jscoverage['protocol/request/Part.js'][19]++;
var PART_HEADER_LENGTH = 16;
_$jscoverage['protocol/request/Part.js'][21]++;
module.exports = Part;
_$jscoverage['protocol/request/Part.js'][23]++;
function Part(options) {
  _$jscoverage['protocol/request/Part.js'][24]++;
  this.kind = options.kind || PartKind.NIL;
  _$jscoverage['protocol/request/Part.js'][25]++;
  this.attributes = options.attributes || 0;
  _$jscoverage['protocol/request/Part.js'][26]++;
  this.argumentCount = options.argumentCount || 0;
  _$jscoverage['protocol/request/Part.js'][27]++;
  this.buffer = undefined;
}
_$jscoverage['protocol/request/Part.js'][31]++;
Part.prototype.toBuffer = (function toBuffer(size) {
  _$jscoverage['protocol/request/Part.js'][32]++;
  var byteLength = util.alignLength(this.buffer.length, 8);
  _$jscoverage['protocol/request/Part.js'][33]++;
  var buffer = new Buffer(PART_HEADER_LENGTH + byteLength);
  _$jscoverage['protocol/request/Part.js'][35]++;
  buffer.writeInt8(this.kind, 0);
  _$jscoverage['protocol/request/Part.js'][37]++;
  buffer.writeInt8(this.attributes, 1);
  _$jscoverage['protocol/request/Part.js'][39]++;
  buffer.writeInt16LE(this.argumentCount, 2);
  _$jscoverage['protocol/request/Part.js'][41]++;
  buffer.writeInt32LE(0, 4);
  _$jscoverage['protocol/request/Part.js'][43]++;
  buffer.writeInt32LE(this.buffer.length, 8);
  _$jscoverage['protocol/request/Part.js'][45]++;
  buffer.writeInt32LE(size, 12);
  _$jscoverage['protocol/request/Part.js'][46]++;
  this.buffer.copy(buffer, PART_HEADER_LENGTH);
  _$jscoverage['protocol/request/Part.js'][47]++;
  if (this.buffer.length < byteLength) {
    _$jscoverage['protocol/request/Part.js'][48]++;
    buffer.fill(0, PART_HEADER_LENGTH + this.buffer.length);
  }
  _$jscoverage['protocol/request/Part.js'][50]++;
  return buffer;
});
_$jscoverage['protocol/request/Part.js'].source = ["// Copyright 2013 SAP AG.","//","// Licensed under the Apache License, Version 2.0 (the \"License\");","// you may not use this file except in compliance with the License.","// You may obtain a copy of the License at","//","// http: //www.apache.org/licenses/LICENSE-2.0","//","// Unless required by applicable law or agreed to in writing,","// software distributed under the License is distributed on an","// \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,","// either express or implied. See the License for the specific","// language governing permissions and limitations under the License.","'use strict';","","var util = require('../../util');","var PartKind = require('../common/PartKind');","","var PART_HEADER_LENGTH = 16;","","module.exports = Part;","","function Part(options) {","  this.kind = options.kind || PartKind.NIL;","  this.attributes = options.attributes || 0;","  this.argumentCount = options.argumentCount || 0;","  this.buffer = undefined;","}","","","Part.prototype.toBuffer = function toBuffer(size) {","  var byteLength = util.alignLength(this.buffer.length, 8);","  var buffer = new Buffer(PART_HEADER_LENGTH + byteLength);","  // Part kind, specifies nature of part data","  buffer.writeInt8(this.kind, 0);","  // Further attributes of part","  buffer.writeInt8(this.attributes, 1);","  // Argument count, number of elements in part data.","  buffer.writeInt16LE(this.argumentCount, 2);","  // Argument count, number of elements in part data (only for some part kinds).","  buffer.writeInt32LE(0, 4);","  // Length of part buffer in bytes","  buffer.writeInt32LE(this.buffer.length, 8);","  // Length in packet remaining without this part.","  buffer.writeInt32LE(size, 12);","  this.buffer.copy(buffer, PART_HEADER_LENGTH);","  if (this.buffer.length &lt; byteLength) {","    buffer.fill(0x00, PART_HEADER_LENGTH + this.buffer.length);","  }","  return buffer;","};"];
