// @sage/syracuse-lib/src/collaboration/entities/copyright.ts
"use strict";
import * as fs from 'fs';
import * as path from 'path';

export const entity = {
	$lockType: "noLock",
	$canDelete: false,
	$canCreate: false,
	$canEdit: false,
	$canSave: false,
	$capabilities: "",
	$titleTemplate: "Credits",
	$descriptionTemplate: "Credits",
	$properties: {
		name: {
			$title: "Name",
		},
		version: {
			$title: "Version",
		},
		license: {
			$title: "Licenses",
		},
		copyright: {
			$title: "Copyrights",
		},
		type: {
			$title: "Type",
			$enum: [{
				$value: "package",
				$title: "Package"
			}, {
				$value: "clientDep",
				$title: "Client dependency"
			}, {
				$value: "component",
				$title: "Component"
			}]
		}
	},
	$functions: {
		$setId(_, context, id) {

		}
	},
	$fetchInstances(_, context, parameters) {
		function createInstance(_, name, version, license, copyright, type) {
			license = license || "";
			copyright = copyright || "";
			var inst = entity.factory.createInstance(_, null, context.db);
			inst.name(_, name);
			inst.version(_, version);
			inst.license(_, license);
			inst.copyright(_, copyright);
			inst.type(_, type);
			return inst;
		}
		var result = [];
		var entity = context.db.model.getEntity(_, "copyright");

		var components = JSON.parse(fs.readFile(path.join(process.cwd(), "copyrights/output-components.json"), "utf8", _));
		components.forEach_(_, function(_, lib) {
			result.push(createInstance(_, lib.name, lib.version, lib.licenses, lib.copyrights, "component"));
		});

		var packs = JSON.parse(fs.readFile(path.join(process.cwd(), "copyrights/output-packages.json"), "utf8", _));
		packs.forEach_(_, function(_, lib) {
			result.push(createInstance(_, lib.name, lib.version, lib.licenses, lib.copyrights, "package"));
		});

		var deps = JSON.parse(fs.readFile(path.join(process.cwd(), "copyrights/output-deps.json"), "utf8", _));
		deps.forEach_(_, function(_, lib) {
			result.push(createInstance(_, lib.name, lib.version, lib.licenses, lib.copyrights, "clientDep"));
		});


		return result;
	}
};
