// @sage/syracuse-lib/src/collaboration/entities/epToUpdate.ts
"use strict";

const locale = require('streamline-locale');
var date = require('@sage/syracuse-core').types.date;

import * as ez from 'ez-streams';
import * as httpClient from '../../../src/http-client/httpClient';
import * as config from 'config'; // must be first syracuse require

function listFolders(_, instance, folders, parent, history) {
	if (!folders) return;

	var list = instance.folders(_);
	folders.forEach_(_, function(_, folder) {
		var item = list.add(_);
		item.parent(_, parent || "");
		item.name(_, folder.name);
		item.release(_, folder.release);
		item.history(_, history || false);
		var folderDate = (folder.dat === "00/00/0000") ? "01/01/1970" : folder.dat;
		item.updated(_, date.parse(folderDate, "dd/MM/yyyy"));
		item.patch(_, folder.nump);
		item.legislations(_, folder.legislations || "");
		listFolders(_, instance, folder.folders, folder.name);
		//if (folder.history) listFolders(_, instance, [folder.history], folder.name, true);
	});
}

export const entity = {
	$isPersistent: false,
	$canSave: false,
	$titleTemplate: "Patches",
	$descriptionTemplate: "Patches",
	$helpPage: "Patches",
	$properties: {},
	$relations: {
		endpoint: {
			$type: "endPoint",
			$title: "Endpoint",
			$isMandatory: true,
			$displayLength: 15,
			$propagate(_, instance, ep) {
				var foldersTree = ep.getService(_, "foldersTree");

				// Reset the list of folders
				instance.folders(_).reset(_);

				if (foldersTree && foldersTree.name) {
					listFolders(_, instance, [foldersTree]);
				}
			},
		},
		folders: {
			$type: "folderToUpdates",
			$title: "Folders to update",
			$isMandatory: false,
			$inv: "epToUpdate",
			$isChild: true,
			$capabilities: "delete",
			$isPlural: true,
			$treeview: {
				$mode: "parentKey",
				$bindings: {
					$id: "name",
					$parent: "parent"
				}
			}
		}
	},
	$functions: {},
	$services: {}
};
