// @sage/syracuse-lib/src/collaboration/entities/friendServer.ts
"use strict";

const locale = require('streamline-locale');
import * as ez from 'ez-streams';
import * as fs from 'fs';
import * as fsp from 'path';
import * as https from 'https';
import * as config from 'config';
import * as certtools from '../../load-balancer/certTools';

export const entity = {
	$titleTemplate: "Collaboration friend server",
	$valueTemplate: "{description}",
	$helpPage: "Administration-reference_friend-servers",
	$properties: {
		description: {
			$title: "Description",
			$isMandatory: true,
			$linksToDetails: true
		},
		endpointUrl: {
			$title: "Collaboration endpoint url",
			$isMandatory: true
		}
	},
	$relations: {
		certificate: {
			$title: "Client certificate",
			$type: "certificate",
			$inv: "friendServers",
		}
	},
	$functions: {
		createWritableStream(_, options) {
			var service = options.service;
			// open a session with trusted server and returns a http writable stream
			var cert = this.certificate(_);
			if (config.x3key && !cert) throw new Error("No certificate");
			if (cert) {
				var opt = {
					url: this.endpointUrl(_) + "/" + service,
					method: options.method || "POST"
				};
				return cert.streamHttpRequest(_, opt);
			} else {
				var certPath = options.certificatesPath;
				var certName = options.certificateName;
				var opt = {
					agent: false,
					key: fs.readFile(fsp.join(certPath, certName + ".key"), _),
					cert: fs.readFile(fsp.join(certPath, certName + ".crt"), _),
					// use our server ca, should be the same
					ca: fs.readFile(fsp.join(__dirname, "../../../syracuse-main/lib/ssl", "ca.crt"), _),
					//
					url: this.endpointUrl(_) + "/" + service,
					method: options.method || "POST"
				};
			}
			return ez.devices.http.client(opt);
		}
	},
	$services: {
		/*	// just for testing friend server without data
  		$createPatch: {
			$method: "POST",
			$isMethod: true,
			$title: "Test",
			$execute(_, context, instance) {
				var stream = instance.createWritableStream(_, {service: "license/test", method: "GET"})
				var resp = stream.end().response(_);
				var respData = resp.readAll(_);
				console.log("Answer "+respData);
				instance.$addDiagnose("success", respData);
			}
		},
	*/
	},
	$events: {
		$canSave: [

			function(_, instance) {
				// there must be at least one active connection
				var certificate = instance.certificate(_);
				if (certificate && certificate.server(_)) {
					instance.$addError(locale.format(module, "clientNotServer"));
				}
				if (certificate && certificate.internal(_)) {
					instance.$addError(locale.format(module, "noInternal"));
				}
				return;
			}
		]
	}
};
