// @sage/syracuse-lib/src/collaboration/entities/licenseModuleData.ts
"use strict";

import * as util from 'util';

export const entity = {
	$titleTemplate: "License module data",
	$valueTemplate: "",
	$descriptionTemplate: "Version",
	$isPersistent: false,
	$canEdit: false,
	$canDelete: false,
	$canCreate: false,
	$listTitle: "List of module data in license",
	$key: "",
	$properties: {
		title: {
			$title: "Title",
			$isReadOnly: true,
			$isLocalized: true
		},
		code: {
			$title: "Code",
			$isReadOnly: true
		},
		availability: {
			$title: "Availability",
			$isReadOnly: true,
			$enum: [{
				$value: "licensed",
				$title: "licensed"
			}, {
				$value: "never",
				$title: "never"
			}, {
				$value: "always",
				$title: "always"
			}, {
				$value: "unlicensed",
				$title: "not licensed"
			}]
		},
		valStart: {
			$title: "Start of validity",
			$type: "date",
			$isReadOnly: true
		},
		valEnd: {
			$title: "End of validity",
			$type: "date",
			$isReadOnly: true
		},
		keyFunctions: {
			$title: "Key functions",
			$isReadOnly: true
		}
	},
	$relations: {},
	$fetchInstances(_, context, parameters) {
		return [];
	},
	$functions: {
		$setId(_, context, id) {
			//
		}
	},
	$links: {}
	// ,
	// $defaultOrder: [["relNumber", true], ["patchNumber", true]]
};
