// @sage/syracuse-lib/src/collaboration/entities/lookupEntityAttr.ts
"use strict";

import { flows } from 'streamline-runtime';
import { helpers } from '@sage/syracuse-core';
import { AdminHelper as adminHelper } from '../../collaboration/helpers';
const locale = require('streamline-locale');

export const entity = {
	$properties: {
		name: {
			$title: "Name"
		},
		title: {
			$title: "Title"
		},
		attrType: {
			$title: "Type"
		}
	},
	$fetchInstances(_, context, parameters) {
		var self = this;
		//
		var preliminary = [];
		var ep = adminHelper.getEndpoint(_, {
			dataset: context.parameters.dataset
		});
		if (!ep) return [];
		//
		var db = ep.getOrm(_);
		var entity = db.getEntity(_, context.parameters.entity);
		if (!entity) return [];
		//
		entity && flows.eachKey(_, entity.$properties, function(_, key, prop) {
			var instance = self.factory.createInstance(_, null, context.db, context);
			instance.name(_, key);
			instance.title(_, prop.$title || key);
			instance.attrType(_, prop.$type);
			preliminary.push(instance);
		});
		entity && flows.eachKey(_, entity.$relations, function(_, key, prop) {
			var instance = self.factory.createInstance(_, null, context.db, context);
			instance.name(_, key);
			instance.title(_, prop.$title || key);
			instance.attrType(_, prop.$type);
			preliminary.push(instance);
		});
		//
		return preliminary;
	},
	$defaultOrder: [
		["name", true]
	]
};
