// @sage/syracuse-lib/src/collaboration/entities/lookupEntity.ts
"use strict";

import * as dataModel from '../../../src/orm/dataModel';
import * as sdataRegistry from '../../../src/sdata/sdataRegistry';
import { flows } from 'streamline-runtime';
import { helpers } from '@sage/syracuse-core';
import { AdminHelper as adminHelper } from '../../collaboration/helpers';

export const entity = {
	$properties: {
		name: {
			$title: "Name"
		},
		title: {
			$title: "Title"
		}
	},
	$fetchInstances(_, context, parameters) {
		var self = this;
		//
		// TODO: dataset ?
		var preliminary = [];
		var application = adminHelper.getApplication(_, context.parameters.application, context.parameters.contract);
		if (!application) return [];
		if (application.protocol(_) === "syracuse") {
			var contract = sdataRegistry.getContract(application.application(_), application.contract(_));
			contract && flows.eachKey(_, contract.entities, function(_, key, entity) {
				var instance = self.factory.createInstance(_, null, context.db, context);
				instance.name(_, helpers.string.pluralize(key));
				instance.title(_, entity.$listTitle || key);
				preliminary.push(instance);
			});
		}
		//
		return preliminary;
	},
	$defaultOrder: [
		["name", true]
	]
};
