// syracuse-ui/lib/tests/charts/spider.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

exports.load = function(page) {
	page.setTitle("HighCharts Spider");
	page.$prototype.$properties = {
		"A": {
			$type: "application/x-array",
			$title: "spiderweb",
			$axes: [{
				$title: "Articles vendus",
				$hierarchies: [
					["article_h"]
				]
			}, {
				$title: "measures",
				$hierarchies: [
					["$measures"]
				]
			}],
			$cube: {
				"$mode": "table",
				"$style": "spiderweb",
				"$displaysOneMeasure": false,
				$title: "Coûts articles",
				$hierarchies: {
					"article_h": {
						$title: "Articles",
						"$properties": ["ARTICLE"],
						"$members": [],
						"$isDefault": true
					}
				},
				$measures: {
					matiere_premiere: {
						"$property": "matiere_premiere",
						$title: "Matière premiere"
					},
					main_oeuvre: {
						"$property": "main_oeuvre",
						$title: "Main d'oeuvre"
					},
					cout_machine: {
						"$property": "cout_machine",
						$title: "Cout machine"
					},
					cout_machine2: {
						"$property": "cout_machine2",
						$title: "Cout machine 2"
					},
					cout_machine3: {
						"$property": "cout_machine3",
						$title: "Cout machine 3"
					}
				}
			},
			"$item": {
				"$properties": {
					"ARTICLE": {
						$title: "Intitulé article",
						$type: "application/x-string"
					},
					"code": {
						$title: "Code article",
						$type: "application/x-string"
					},
					matiere_premiere: {
						$title: "Matiere première",
						$type: "application/x-decimal"
					},
					main_oeuvre: {
						$title: "Main d'oeuvre",
						$type: "application/x-decimal"
					},
					cout_machine: {
						$title: "Coût machine",
						$type: "application/x-decimal"
					},
					cout_machine2: {
						$title: "Coût machine 2",
						$type: "application/x-decimal"
					},
					cout_machine3: {
						$title: "Coût machine 3",
						$type: "application/x-decimal"
					}
				},
				"$links": {
					"$details": {
						$title: "Details",
						"$url": "{$baseUrl}/article('{$key}')",
						$type: "{$baseType}.article.$details"
					},
					"$edit": {
						$title: "Edit",
						"$url": "{$url}/$workingCopies",
						$type: "{$baseType}.article.$edit",
						"$method": "POST"
					},
					"$delete": {
						$title: "Delete",
						"$url": "{$url}",
						$type: null,
						"$method": "DELETE"
					}
				}
			}
		},
		"B": {
			"$type": "application/x-array",
			"$title": "column, $isStacked",
			"$axes": [{
				"$title": "Articles vendus",
				"$hierarchies": [
					["article_h"]
				]
			}, {
				"$title": "measures",
				"$hierarchies": [
					["$measures"]
				]
			}],
			"$cube": {
				"$mode": "table",
				"$style": "column",
				"$displaysOneMeasure": false,
				"$title": "Coûts articles",
				"$hierarchies": {
					"article_h": {
						"$title": "Articles",
						"$properties": ["ARTICLE"],
						"$members": [],
						"$isDefault": true
					}
				},
				"$measures": {
					"matiere_premiere": {
						"$property": "matiere_premiere",
						"$title": "Matière premiere",
						"$isStacked": true,
						"$isNormalized": false,
						"$stackingGroup": 0
					},
					"main_oeuvre": {
						"$property": "main_oeuvre",
						"$title": "Main d'oeuvre",
						"$isStacked": true,
						"$isNormalized": false,
						"$stackingGroup": 0
					},
					"cout_machine": {
						"$property": "cout_machine",
						"$title": "Cout machine",
						"$isStacked": true,
						"$isNormalized": false,
						"$stackingGroup": 0
					}
				}
			},
			"$item": {
				"$properties": {
					"ARTICLE": {
						"$title": "Intitulé article",
						"$type": "application/x-string"
					},
					"code": {
						"$title": "Code article",
						"$type": "application/x-string"
					},
					"matiere_premiere": {
						"$title": "Matiere première",
						"$type": "application/x-decimal"
					},
					"main_oeuvre": {
						"$title": "Main d'oeuvre",
						"$type": "application/x-decimal"
					},
					"cout_machine": {
						"$title": "Coût machine",
						"$type": "application/x-decimal"
					}
				},
				"$links": {
					"$details": {
						"$title": "Details",
						"$url": "{$baseUrl}/article('{$key}')",
						"$type": "{$baseType}.article.$details"
					},
					"$edit": {
						"$title": "Edit",
						"$url": "{$url}/$workingCopies",
						"$type": "{$baseType}.article.$edit",
						"$method": "POST"
					},
					"$delete": {
						"$title": "Delete",
						"$url": "{$url}",
						"$type": null,
						"$method": "DELETE"
					}
				}
			}
		},
		"C": {
			"$type": "application/x-array",
			"$title": "column, $isStacked, $isNormalized",
			"$axes": [{
				"$title": "Articles vendus",
				"$hierarchies": [
					["article_h"]
				]
			}, {
				"$title": "measures",
				"$hierarchies": [
					["$measures"]
				]
			}],
			"$cube": {
				"$mode": "table",
				"$style": "column",
				"$displaysOneMeasure": false,
				"$title": "Coûts articles",
				"$hierarchies": {
					"article_h": {
						"$title": "Articles",
						"$properties": ["ARTICLE"],
						"$members": [],
						"$isDefault": true
					}
				},
				"$measures": {
					"matiere_premiere": {
						"$property": "matiere_premiere",
						"$title": "Matière premiere",
						"$isStacked": true,
						"$isNormalized": true,
						"$stackingGroup": 0
					},
					"main_oeuvre": {
						"$property": "main_oeuvre",
						"$title": "Main d'oeuvre",
						"$isStacked": true,
						"$isNormalized": true,
						"$stackingGroup": 0
					},
					"cout_machine": {
						"$property": "cout_machine",
						"$title": "Cout machine",
						"$isStacked": true,
						"$isNormalized": true,
						"$stackingGroup": 0
					}
				}
			},
			"$item": {
				"$properties": {
					"ARTICLE": {
						"$title": "Intitulé article",
						"$type": "application/x-string"
					},
					"code": {
						"$title": "Code article",
						"$type": "application/x-string"
					},
					"matiere_premiere": {
						"$title": "Matiere première",
						"$type": "application/x-decimal"
					},
					"main_oeuvre": {
						"$title": "Main d'oeuvre",
						"$type": "application/x-decimal"
					},
					"cout_machine": {
						"$title": "Coût machine",
						"$type": "application/x-decimal"
					}
				},
				"$links": {
					"$details": {
						"$title": "Details",
						"$url": "{$baseUrl}/article('{$key}')",
						"$type": "{$baseType}.article.$details"
					},
					"$edit": {
						"$title": "Edit",
						"$url": "{$url}/$workingCopies",
						"$type": "{$baseType}.article.$edit",
						"$method": "POST"
					},
					"$delete": {
						"$title": "Delete",
						"$url": "{$url}",
						"$type": null,
						"$method": "DELETE"
					}
				}
			}
		},
		"D": {
			"$type": "application/x-array",
			"$title": "column, $isStacked, $isNormalized, $stackingGroup",
			"$axes": [{
				"$title": "Articles vendus",
				"$hierarchies": [
					["article_h"]
				]
			}, {
				"$title": "measures",
				"$hierarchies": [
					["$measures"]
				]
			}],
			"$cube": {
				"$mode": "table",
				"$style": "column",
				"$displaysOneMeasure": false,
				"$title": "Coûts articles",
				"$hierarchies": {
					"article_h": {
						"$title": "Articles",
						"$properties": ["ARTICLE"],
						"$members": [],
						"$isDefault": true
					}
				},
				"$measures": {
					"matiere_premiere": {
						"$property": "matiere_premiere",
						"$title": "Matière premiere",
						"$isStacked": true,
						"$isNormalized": true,
						"$stackingGroup": 1
					},
					"main_oeuvre": {
						"$property": "main_oeuvre",
						"$title": "Main d'oeuvre",
						"$isStacked": true,
						"$isNormalized": true,
						"$stackingGroup": 0
					},
					"cout_machine": {
						"$property": "cout_machine",
						"$title": "Cout machine",
						"$isStacked": true,
						"$isNormalized": true,
						"$stackingGroup": 0
					}
				}
			},
			"$item": {
				"$properties": {
					"ARTICLE": {
						"$title": "Intitulé article",
						"$type": "application/x-string"
					},
					"code": {
						"$title": "Code article",
						"$type": "application/x-string"
					},
					"matiere_premiere": {
						"$title": "Matiere première",
						"$type": "application/x-decimal"
					},
					"main_oeuvre": {
						"$title": "Main d'oeuvre",
						"$type": "application/x-decimal"
					},
					"cout_machine": {
						"$title": "Coût machine",
						"$type": "application/x-decimal"
					}
				},
				"$links": {
					"$details": {
						"$title": "Details",
						"$url": "{$baseUrl}/article('{$key}')",
						"$type": "{$baseType}.article.$details"
					},
					"$edit": {
						"$title": "Edit",
						"$url": "{$url}/$workingCopies",
						"$type": "{$baseType}.article.$edit",
						"$method": "POST"
					},
					"$delete": {
						"$title": "Delete",
						"$url": "{$url}",
						"$type": null,
						"$method": "DELETE"
					}
				}
			}
		}
	};
	page.leftCol.style.width = page.rightCol.style.width = "50%";
	page.loadNewItem(page.leftCol, {
		$bind: "A",
		$renderer: "highCharts"
	});
	page.loadNewItem(page.leftCol, {
		$bind: "B",
		$renderer: "highCharts"
	});
	page.loadNewItem(page.rightCol, {
		$bind: "C",
		$renderer: "highCharts"
	});
	page.loadNewItem(page.rightCol, {
		$bind: "D",
		$renderer: "highCharts"
	});
	var data = [{
		$uuid: "d3e97372-21c1-40ac-b147-4a0eaad4f58e",
		$key: "d3e97372-21c1-40ac-b147-4a0eaad4f58e",
		"ARTICLE": "Toto",
		"code": "TOOT",
		matiere_premiere: "10.3",
		main_oeuvre: "20.5",
		cout_machine: "56.3",
		cout_machine2: "46.3",
		cout_machine3: "60.3"
	}, {
		$uuid: "a3e97372-21c1-40ac-b147-4a0eaad4f58e",
		$key: "a3e97372-21c1-40ac-b147-4a0eaad4f58e",
		"ARTICLE": "Cle USB",
		"code": "CLEUSB",
		matiere_premiere: "12.3",
		main_oeuvre: "45.2",
		cout_machine: "125.2",
		cout_machine2: "120.2",
		cout_machine3: "130.3"
	}, {
		$uuid: "b3e97372-21c1-40ad-b147-4a0eaad4f58e",
		$key: "b3e97372-21c1-40ad-b147-4a0eaad4f58e",
		"ARTICLE": "Ecran Plasma",
		"code": "ECRPLASMA",
		matiere_premiere: "125.2",
		main_oeuvre: "50.3",
		cout_machine: "100.1",
		cout_machine2: "90.1",
		cout_machine3: "120.3"
	}, {
		$uuid: "c3e97372-21c2-40ac-b147-4a0eaad4f58e",
		$key: "c3e97372-21c2-40ac-b147-4a0eaad4f58e",
		"ARTICLE": "Lecteur DVD",
		"code": "Lecteur DVD",
		matiere_premiere: "45.2",
		main_oeuvre: "23.3",
		cout_machine: "22.3",
		cout_machine2: "12.3",
		cout_machine3: "32.3"
	}, {
		$uuid: "d4e97342-21c1-40ac-b147-4a0eaad4f58e",
		$key: "d4e97342-21c1-40ac-b147-4a0eaad4f58e",
		"ARTICLE": "Imprimante Laser",
		"code": "IMPLASER",
		matiere_premiere: "45.2",
		main_oeuvre: "65.00",
		cout_machine: "45.25",
		cout_machine2: "40.25",
		cout_machine3: "50.25"
	}, {
		$uuid: "d5e97372-22c1-40ac-b147-4a0eaad4f58e",
		$key: "d5e97372-22c1-40ac-b147-4a0eaad4f58e",
		"ARTICLE": "CD-ROM",
		"code": "CDROM",
		matiere_premiere: "200.0",
		main_oeuvre: "120.32",
		cout_machine: "100.5",
		cout_machine2: "200.3",
		cout_machine3: "120.3"
	}];
	page.applyChange({
		"A": helpers.object.clone(data, true),
		"B": helpers.object.clone(data, true),
		"C": helpers.object.clone(data, true),
		"D": helpers.object.clone(data, true),
	});
};
