// syracuse-ui/lib/field/array/singleBuilder/singleBuilder.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var SingleRecord = require('./singleRecord').SingleRecord;
var SelectorPopup = require('syracuse-ui/lib/site/selectorPopup').SelectorPopup;

function SingleBuilder(){
}

exports.SingleBuilder = helpers.defineClass(SingleBuilder, null, {
    initialize: function(){
        this.list.$item.$isQuickDesignerEnabled = false;
        this.list.$item.$isPagerHidden = true;
        this.RecordClass = SingleRecord;
        this.isSingleBuilder = true;
    },
    applyDesignMetaData: function(metaData, onAuthoring){
    },
    emptyBody: function(addEmptySlot){
        $(this._body).empty();
    },
    drawBuilder: function(){
        this.list.$skin = this.list.$item.$skin || "s-array";
        this._body = document.createElement("div");
        this._body.className = this.list.$skin + "-body";
        this.list._core.appendChild(this._body);
        this.list._topbar.style.display = "none";
        this.list.applyDesignMetaData(this.list.$item, false);
        this.list.parseCapabilities(this.list.$prototype.$capabilities);
        
        this.list.setState(this.list.$field);
        this.list.boxParent.getArticle().bind(this.list, this.list.$item.$bind);
        
        if (this.list.$prototype.$item.$type == "application/x-choice") {
            if (!(this.list.$prototype.$links && this.list.$prototype.$links.$select)) {
                this.list.applyChange({
                    $links: {
                        $select: {
                            $title: this.list._localize.slist_select
                        }
                    }
                });
            }            
        }
        this.list.applyCapabilities();
        if (this.list._contextMenu) {
            this.list._topbar.style.display = "";
        }
    },
    onAfterDataBinding: function(){
        if (this.list.$isEditMode) {
            var store = this.list._store;
            if (this.list.$prototype.$minItems) {
                while (store._records.length < this.list.$prototype.$minItems) {
                    store.appendRecord({
                        $recordIndex: store._records.length
                    });
                }
            }
            store.checkMaxItems();
        }
    },
    dispose: function(){
        delete this.$layoutOptions;
        delete this.list;
    },
    loadChoiceSelector: function(){
        var self = this;
        if (!self._popup) {
            (self._popup = new SelectorPopup()).open(self.list, {
                $properties: {
                    $resources: {
                        $type: "application/x-array",
                        $item: {
                            $type: "application/json",
                            $properties: {
                                $title: {
                                    $title: "description",
                                    $type: "application/x-string",
                                    $capabilities: "sort"
                                }
                            }
                        }
                    }
                },
                $resources: self.list.$prototype.$item.$value.$enum.map(function($choice){
                    return {
                        $uuid: $choice.$value,
                        $title: $choice.$title
                    };
                }),
                $$position: self.list.menuItems.$select[0].$$item,
                onClose: function(){
                    delete self._popup;
                },
                onValidate: function(){
                    self.list._store.notifySelectRecords(self.list.currentSelectRecords);
                },
                onSelectRecord: function(selectedRecords){
                    self.list.currentSelectRecords = selectedRecords;
                    return false;
                }
            });
        }
        else {
            if (self._popup) {
                self._popup.close();
            }
        }
    }
});
