// syracuse-x3/lib/convergence/model/CvgClose.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var CvgObservable = require('syracuse-x3/lib/convergence/model/CvgObservable').CvgObservable;

var CvgClose = exports.CvgClose = helpers.defineClass(function(aClient, reason, exception, appCode) {
	CvgObservable.call(this, aClient);
	this.closeReason = reason;
	this.closeMessage = exception && exception.message || "";
	this.closeAppCode = appCode;
	this.closeStack =  exception ? require('util').inspect(exception.stack) : "";
	console.log(this.closeStack);

	// En dernier
	this.setJsonType('SESSION');
}, CvgObservable, {
	toJSON: function() {
		
//		1: confirmation fermeture explicite utilisateur
//		2: time-out inactivité
//		3: fermeture à l'initiative du back office
//		4: fermeture forcée serveur back office à la connexion
//		101: pb communication avec back office
//		102: plantage serveur web
		var close = {
			reason: this.closeReason,
		};
		if (close.reason > 100) {
			close.$diagnoses = [{
				message: this.closeMessage,
				stack: this.closeStack,
				appCode: this.closeAppCode,
			}]
		}
		
		return {close: close};
	},
	toJsonObj: function() {
		return toJSON();
	}
});
