// syracuse-x3/lib/convergence/model/CvgClose.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var CvgObservable = require('syracuse-x3/lib/convergence/model/CvgObservable').CvgObservable;

var CvgClose = helpers.defineClass(function(aClient, reason, ex, appCode) {
	CvgObservable.call(this, aClient);
	this.closeReason = reason;
	this.closeAppCode = appCode;
	this.excep = ex;
	// Clean blobs from SvcBlob
	this.client.services.blobSvc.blobs = {};
	// En dernier
	this.setJsonType('SESSION');
}, CvgObservable, {
	toJSON: function() {

		//		1: confirmation fermeture explicite utilisateur
		//		2: time-out inactivité
		//		3: fermeture à l'initiative du back office
		//		4: fermeture forcée serveur back office à la connexion
		//		101: pb communication avec back office
		//		102: plantage serveur web
		var close = {
			reason: this.closeReason,
		};
		if (close.reason > 100) {
			close.$diagnoses = [{
				message: this.excep ? (this.excep.message ? this.excep.message : this.excep) : "",
				stack: this.excep && this.excep.stack ? require('util').inspect(this.excep.stack) : "",
				appCode: this.closeAppCode,
			}];
		}
		return {
			close: close
		};
	},
	toJsonObj: function() {
		return this.toJSON();
	}
});

// Factory pattern avoids jslint/jshint complaining about using "new" for side effects
exports.create = function(aClient, reason, ex, appCode) {
	return new CvgClose(aClient, reason, ex, appCode);
};
