// syracuse-x3/lib/convergence/model/CvgDiagnose.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var CvgObservable = require('syracuse-x3/lib/convergence/model/CvgObservable').CvgObservable;

var tracerStd = require("syracuse-trace/lib/helper").getTracer("classic.std");

var _log = true;
var CvgDiagnose = helpers.defineClass(function(client, severity, appCode, message, details, error, origin, recovery, track, jsonT) {
	CvgObservable.call(this, client);
	this.severity = severity;
	var errMsg = null;
	if (error) {
		errMsg = error.message || error;
	}
	this.message = message || errMsg;

	this.error = error;
	this.origin = origin;
	if (details && details.length > 0) {
		this.details = details;
	} else if (errMsg && this.message !== errMsg) {
		this.details = errMsg;
	}
	// #1231 - Force abort if client no more alive
	if (client && client.isAlive()) {
		this.appCode = appCode;
	} else {
		this.appCode = 500;
	}
	this.recovery = recovery || 'KILL';
	// En dernier
	if (track) {
		this.setJsonType(jsonT || 'DIAGNOSE');
	}
}, CvgObservable, {
	toJSON: function() {
		var diag = {
			"$message": this.message,
			"$severity": this.severity || 'ERROR',
			"appCode": this.appCode || 500,
			"$origin": "NodeJS: Convergence Server" + (this.origin ? (": " + this.origin) : ""),
			"$stackTrace": this.error ? this.error.stack : null,
			"$diagnoses": this.$diagnoses || null
		};
		if (this.details) {
			diag.$details = this.details;
		}
		if (this.recovery) {
			diag.$recoveryCode = this.recovery;
		}
		diag.$links = [{
			'$title': "Sage ERP X3 support web site",
			'$url': "http://www.sage.com/support"
		}];
		return diag;
	},
	// Filter same diagnoses - send only one (ex: bad format in lists)
	sendToClient: function(trackList) {
		if (trackList == null) {
			return true;
		}
		var i, d;
		for (i = 0; i < trackList.length; i++) {
			d = trackList[i];
			if (d.message === this.message && d.details === this.details && d.origin === this.origin) {
				if ((!d.error && !this.error) || (d.error && this.error && d.error.message && this.error.message)) {
					return false;
				}
			}
		}

		if (_log) tracerStd.warn() && tracerStd.warn((new Date().toISOString()) + " CvgDiagnose -> " + this.message + (this.error ? "\n\t" + this.error.message : "") + (this.details ? "\n\t" + this.details : ""));
		return true;
	}

});

// Factory pattern avoids jslint/jshint complaining about using "new" for side effects
exports.create = function(client, severity, appCode, message, details, error, origin, recovery, track, jsonT) {
	return new CvgDiagnose(client, severity, appCode, message, details, error, origin, recovery, track, jsonT);
};
