// syracuse-x3/lib/convergence/model/CvgDiagnose.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var CvgObservable = require('syracuse-x3/lib/convergence/model/CvgObservable').CvgObservable;

var _tempAlreadySent=[];
var _log=true;
var CvgDiagnose = exports.CvgDiagnose = helpers.defineClass(function(client, severity, appCode, message, details, error, origin, recovery, track) {
	CvgObservable.call(this, client);
	this.severity = severity;
	this.appCode = appCode;
	this.message = message;
	this.error = error;
	this.origin = origin;
	this.details = details;
	this.recovery = recovery;
	if(_log){
		var x="CvgDiagnose -> " + message +  (error ? "\n\t" + error.message : "")  +  (details ? "\n\t" + details : "");
		if (_tempAlreadySent.indexOf(x)==-1) {
			// Temporay - Avoid multiple same diagnoses
			_tempAlreadySent.push(x);
			console.log(x);
		}
	}
	// En dernier
	if (track)
		this.setJsonType('DIAGNOSE');
	
}, CvgObservable, {

	toJSON: function() {
		var diag = {
			message: this.message,
			severity: this.severity || 'ERROR',
			appCode: this.appCode || 500,
			$origin: "NodeJS: Convergence Server" + (this.origin ? (": "+this.origin) : ""),
			
		}
		if (this.details)
			diag.$details = this.details;
		
		if (this.error) {
			diag.$diagnoses = [{
				message: this.error.message,
				stackTrace: this.error.stack
	        }];
		}
		
        if (this.recovery) {
        	diag.$recoveryCode = this.recovery;
        }
        
		diag.$links = [{
           title: "Sage ERP X3 support web site",
           $url: "http://www.sage.com/support"
        }];
		return diag;
	},

});
