// syracuse-x3/lib/convergence/model/CvgException.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var CvgDiagnose = require('syracuse-x3/lib/convergence/model/CvgDiagnose').CvgDiagnose;

var CustomException = helpers.defineClass(function(client, e, message) {
	Error.call(this);
	Error.captureStackTrace(this, this); //super helper method to include stack trace in error object
	this.name = 'Custom Exception'; //set our function’s name as error name.
	this.message = e && e.message ?  e.message : (message ? message : ""); //set the error message
	this.inner = e;
	
	
}, Error, {
});
	
var CvgException = exports.CvgException = helpers.defineClass(function(client, e, message, recoveryCode, origin) {
	CustomException.call(this, client, e, message);
	this.name = 'CvgException'; //set our function’s name as error name.
	// Create diagnose exception (not tracked by the client !!)
	this.diagnose = new CvgDiagnose(client, 'ERROR', 500, message || this.message, null, this, origin, recoveryCode || 'KILL', false)
}, CustomException, {
});


var CvgFormatException = exports.CvgFormatException = helpers.defineClass(function(client, e, message) {
	CustomException.call(this, client, e, message);
	this.name = 'CvgFormatException'; //set our function’s name as error name.
	// Create diagnose exception (tracked by the client !!)
	this.diagnose = new CvgDiagnose(client, 'ERROR', 500, message || this.message, null, this, "Format", null, false)
}, CustomException, {
});
