// syracuse-x3/lib/convergence/model/CvgException.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var CvgDiagnose = require('syracuse-x3/lib/convergence/model/CvgDiagnose').CvgDiagnose;

var CustomException = helpers.defineClass(function(client, e, s) {
	Error.call(this);
	Error.captureStackTrace(this, this); //super helper method to include stack trace in error object
	this.name = 'Custom Exception'; //set our function’s name as error name.
	this.message = e && e.message || s; //set the error message
	this.inner = e;
}, Error, {
});
	
var CvgException = exports.CvgException = helpers.defineClass(function(client, e, s) {
	CustomException.call(this, client, e, s);
	this.name = 'CvgException'; //set our function’s name as error name.
	// Create diagnose exception (not tracked by the client !!)
	this.diagnose = new CvgDiagnose(client, 'ERROR', 'ERROR', (e && e.message) || s, e && e.stack, true)
}, CustomException, {
});


var CvgFormatException = exports.CvgFormatException = helpers.defineClass(function(client, e, s) {
	CustomException.call(this, client, e, s);
	this.name = 'CvgFormatException'; //set our function’s name as error name.
	// Create diagnose exception (tracked by the client !!)
	this.diagnose = new CvgDiagnose(client, 'ERROR', 'FORMAT', (e && e.message) || s, e && e.stack, false)
}, CustomException, {
});
