// syracuse-x3/lib/convergence/model/CvgInstanceExeci.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var CvgObservable = require('syracuse-x3/lib/convergence/model/CvgObservable').CvgObservable;

var CvgInstanceExeci = exports.CvgInstanceExeci = helpers.defineClass(function(ist,spgm,dbgflag,paramDefs,paramVal, client) {
	CvgObservable.call(this, client);
	this.instance = ist;
	this.spgm = spgm;
	this.dbgflag = dbgflag;
	this.defs = paramDefs;
	this.values = paramVal
	this.xml = "";
	// En dernier
	this.setJsonType('TARGET');
}, CvgObservable, {
	format: {
		get: function() {
			return this.instance == null ? "" : this.instance.format;
		}
	},
	getNl: function() {
		return 0;
	},
	getXid: function() {
		return this.instance == null ? "" : (this.instance.ist.blocAlphaId + this.instance.ist.fieldId);
	},
	hasFormat: function() {
		return this.instance == null ? false : this.instance.hasFormat();
	},
	setInstance: function(instance) {
		this.instance = instance;
	},
	setXml: function(xml) {
		this.xml = xml;
	},
	toJSON: function() {
		var object = {};
		object.type = 'rpc';
		var rpc = {};
		rpc.xml = this.xml;
		object['rpc'] = rpc;
		var ist = {};
		if (this.instance != null) {
			ist.xid = this.getXid();
			if (this.instance.hasFormat()) {
				ist.fmt = this.format;
			}
			ist.win = this.instance.ist.winAlphaId;
			ist.edit = true;
			rpc.ist = ist;
		}
		return object;
	}
});
