// syracuse-x3/lib/convergence/model/CvgMessage.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var CvgObservable = require('syracuse-x3/lib/convergence/model/CvgObservable').CvgObservable;
var CAdonixMessage = Object; //require('syracuse-x3/lib/convergence/CAdonixMessage').CAdonixMessage;

var CvgMessageObersable = exports.CvgMessageObersable = helpers.defineClass(function(message, client) {
	CvgObservable.call(this, client);
	this.message = message;
	this.jsonType = 'TARGET'; // mess superviseur/X3
	this.forceChange();
}, CvgObservable, {
	toJSON: function() {
		var target = {};
		target.type = "box";
		var box = {};

		box.type = this.message.type;
		box.def = this.message.defBtn;
		// box.put("to",get());
		box.tit = this.message.title;
		var li = [];
		li.push(this.message.message);
		box.li = li;
		box.to = this.message.sleep;
		target.box = box;
		return target;
	}

});

var ABORT = 99;
var WEBSERVER = 98;


var CvgMessage = exports.CvgMessage = helpers.defineClass(function(aType, aAction, aMessage, aTitle, aDefBtn, aSleep, aChx, clientObs, notify) {
	CAdonixMessage.call(this, aType, aMessage);
	this.action = aAction;
	this.webServerLogicalMess = null;
	this.clientObj = null;
	this.title = aTitle;
	this.message = aMessage;
	this.defBtn = aDefBtn;
	this.sleep = aSleep;
	this.chx = aChx;
	this.type = aType;
	this.replyAuto = aType === ABORT || aType === WEBSERVER ? true : false;
	if (clientObs != null && (notify == null || notify)) {
		new CvgMessageObersable(this, clientObs);
	}
}, CAdonixMessage, {
	// TODO - Non utilisé - private CvgMessageObersable observableMessage = null;
	// TODO - Non utilisé - private String this.lib = null;
	// TODO - Non utilisé - private String this.progressMax = null;
	// TODO - Non utilisé - private String this.progressValue = null;
	// TODO - Non utilisé - private String this.vLC_LibNum = null;
	// TODO - Non utilisé - private String this.vLC_LibSuite = null;
	// TODO - Non utilisé - private String this.vLC_LibText = null;
	// TODO - Non utilisé - private String this.vLC_ValNum = null;
	// TODO - Non utilisé - private String this.vLC_ValText = null;
/* TODO overload
	public CvgMessage(aType, aAction, aMessage, aTitre, aDefBtn, aSleep, aChx, clientObs) {
		CAdonixMessage.call(this, aType, aMessage, aTitre, aDefBtn, aSleep, aChx);
		this.action = aAction;
		if (aType == Cvg.AFF_SYNCHRONE_TYPE_VALIDCAR && this.action == Cvg.AFF_SYNCHRONE_ACTION_OPEN) {
			setVlcLibs(aMessage, aTitre);
		} else if (aType == Cvg.AFF_SYNCHRONE_TYPE_VALIDCAR && this.action == Cvg.AFF_SYNCHRONE_ACTION_MODIFY) {
			setVlcValues(aMessage, aTitre);
		} else if (aType == Cvg.AFF_SYNCHRONE_TYPE_PROGRESS && this.action == Cvg.AFF_SYNCHRONE_ACTION_OPEN) {
			setProgressLibs(aTitre);
		} else if (aType == Cvg.AFF_SYNCHRONE_TYPE_PROGRESS && this.action == Cvg.AFF_SYNCHRONE_ACTION_MODIFY) {
			setProgressValues(aMessage);
		}
		if (clientObs != null) {
			new CvgMessageObersable(this, clientObs);
		}
	}*/

	addWebServerLogicalMess: function(aListOfLogicalMess) {
		if (aListOfLogicalMess != null) {
			if (this.webServerLogicalMess == null) {
				this.webServerLogicalMess = [];
			}
			this.webServerLogicalMess.push(aListOfLogicalMess);
		}
	},
	/* TODO overload
	addWebServerLogicalMess: function(aLogicalMess) {
		if (aLogicalMess != null) {
			if (this.webServerLogicalMess == null) {
				this.webServerLogicalMess = [];
			}
			this.webServerLogicalMess.push(aLogicalMess);
		}
	},*/
	clone: function() {
		return new CvgMessage(this.typ, this.action, this.mess, this.title, this.btn, this.sleep, this.chx, this.clientObs);
	},
	isAffSynchroneDebug: function() {
		return this.typ == Cvg.AFF_SYNCHRONE_TYPE_DEBUGON || this.typ == Cvg.AFF_SYNCHRONE_TYPE_DEBUGOFF;
	},
	isErrorMessage: function() {
		return this.typ == CAdonix.MSG_ERROR;
	},
	isStrongMess: function() {
		// Syracuse TODO : 
		var wTypMess = TODO_SUPER.typ;
		return wTypMess == CAdonix.MSG_ERROR || wTypMess == CAdonix.MSG_WARNING;
	},
	setProgressLibs: function(aTitre) {
		this.title = aTitre;
	},
	setProgressValues: function(aMessage) {
		var wST = new StringTokenizer(aMessage, "^");
		if (wST.countTokens() >= 2) {
			// TODO - Non utilisé - this.progressMax = wST.nextToken();
			// TODO - Non utilisé - this.progressValue = wST.nextToken();
			if (wST.hasMoreTokens()) {
				// TODO - Non utilisé - this.lib = wST.nextToken();
			} else {
				// TODO - Non utilisé - this.lib = EMPTY;
			}
		} else {
			// TODO - Non utilisé - this.lib = 'ERR_NOWMAX';
		}
	},
	setVlcLibs: function(aMessage, aTitre) {
		var wST = new StringTokenizer(aTitre, "^");
		if (wST.countTokens() == 3) {
			this.title = wST.nextToken();
			// TODO - Non utilisé - this.lib = wST.nextToken();
			// TODO - Non utilisé - this.vLC_LibSuite = wST.nextToken();
		}

		wST = new StringTokenizer(aMessage, "^");
		if (wST.countTokens() == 2) {
			// TODO - Non utilisé - this.vLC_LibText = wST.nextToken();
			// TODO - Non utilisé - this.vLC_LibNum = wST.nextToken();
		}

	},
	setVlcValues: function(aMessage, aTitre) {
		var wST = new StringTokenizer(aMessage, "^");
		if (wST.countTokens() == 2) {
			// TODO - Non utilisé - this.vLC_ValText = wST.nextToken();
			// TODO - Non utilisé - this.vLC_ValNum = wST.nextToken();
		}
	},
	traverseChilds: function(aTraverser) {
		var wTraversingState = false;
		var wLogicalMessage;
		var wMax = this.webServerLogicalMess != null ? this.webServerLogicalMess.length : 0;
		var wI = 0;
		while (wI < wMax) {
			wLogicalMessage = this.webServerLogicalMess[wI];
			wTraversingState = aTraverser.doWhenTraverse(wLogicalMessage, wI);
			if (wTraversingState == true) {
				return wTraversingState;
			}
			wI++;
		}
		return wTraversingState;
	}
});
