// syracuse-x3/lib/convergence/model/CvgMenuCtx.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var CvgObservable = require('syracuse-x3/lib/convergence/model/CvgObservable').CvgObservable;

var CvgMenuCtx = helpers.defineClass(function(client, id, enabled, label) {
	CvgObservable.call(this, client);
	this.id = id;
	this.enabled = enabled;
	this.label = label;
	// En dernier
	this.setJsonType('TARGETACT');
}, CvgObservable, {

	// Vidage des contextButtons sur noaid NDOMNCO
	clear: function() {
		this.cleared = true;
	},
	getObservableId: function() {
		return this.id.toString();
	},
	toJSON: function() {
		if (this.cleared) {
			return null;
		}

		var object = [this.id.toString(), this.enabled];
		if (this.label) {
			object.push(this.label);
		}
		return object;
	},
	toString: function() {
		return JSON.stringify(this.toJSON());
	}
});

// Factory pattern avoids jslint/jshint complaining about using "new" for side effects
exports.create = function(client, id, enabled, label) {
	return new CvgMenuCtx(client, id, enabled, label);
};
