// syracuse-x3/lib/convergence/model/CvgPluginTarget.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var CvgObservable = require('syracuse-x3/lib/convergence/model/CvgObservable').CvgObservable;
var CvgEntity = require('syracuse-x3/lib/convergence/model/CvgEntity').CvgEntity;
var CvgScreenDummy = require('syracuse-x3/lib/convergence/model/CvgListOfScreens').CvgScreenDummy;
var CvgVariable = require('syracuse-x3/lib/convergence/model/CvgVariable').CvgVariable;
var CvgIst = require('syracuse-x3/lib/convergence/types/CvgIst').CvgIst;

var CvgPluginTarget = exports.CvgPluginTarget = helpers.defineClass(function(cli, name, method, methodid, type, id, module, klass) {
	CvgObservable.call(this, cli);

	this.name = name;
	this.method = method;
	this.type = type;
	this.id = id;
	this.module = module;
	this.class = klass;
	this.methodid = methodid;

	this.setJsonType('TARGET');
}, CvgObservable, {

	toJSON: function() {
		var plug = {
			"class": this.class.toUpperCase(),
			name: this.name,
			type: this.type,
			id: this.id,
			module: this.module,
			method: this.method.toLowerCase(),
			methodid: this.methodid
		};

		return {
			plugin: plug
		};
	}
});
CvgPluginTarget.create = function(client, pluginCall, id) {
	return new CvgPluginTarget(client, pluginCall.name, pluginCall.method, pluginCall.methodid, pluginCall.type, id, pluginCall.module, pluginCall.class);
};
