// syracuse-x3/lib/convergence/model/CvgReplyAckCall.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var CvgObservable = require('syracuse-x3/lib/convergence/model/CvgObservable').CvgObservable;

var CvgReplyAckCAll = exports.CvgReplyAckCall = helpers.defineClass(function(client, spgmId) {
	CvgObservable.call(this, client);
	this.spgmId = spgmId;
	this.reply = false;
//	this.uuid = helpers.uuid.generate();
	
	// En dernier
	this.setJsonType('RPC');
}, CvgObservable, {
//	getObservableId: function() {
//		console.error("UUID: "+this.uuid);
//		return this.uuid;
//	},
	setReply: function(errn, errl, errp, errt, errm, resu, arrayOfParam, arrayOfResu) {
		
		this.errn = errn;
		this.errl = errl;
		this.errp = errp;
		this.errt = errt;
		this.errm = errm;
		this.resu = resu;
		this.arrayOfParam = arrayOfParam;
		this.arrayOfResu = arrayOfResu;
		
		this.reply = true;
		
		
//		console.log("this.errn: "+JSON.stringify(this.errn,null,2));
//		console.log("this.errl: "+JSON.stringify(this.errl,null,2));
//		console.log("this.errp: "+JSON.stringify(this.errp,null,2));
//		console.log("this.errt: "+JSON.stringify(this.errt,null,2));
//		console.log("this.errm: "+JSON.stringify(this.errm,null,2));
//		console.log("this.resu: "+JSON.stringify(this.resu,null,2));
//		console.log("this.arrayOfParam: "+JSON.stringify(this.arrayOfParam,null,2));
//		console.log("this.arrayOfResu: "+JSON.stringify(this.arrayOfResu,null,2));
	},
	toJSON: function() {
		
		var result = {};
		if (this.resu != null) {
			result.typ = this.resu.type;
			result.resu = this.resu.value;
		}
		
		var params = [];
		for (var i in this.arrayOfParam) {
			var param = {};
			if (this.arrayOfParam[i].num != null)
				param.num = this.arrayOfParam[i].num;
			
			if (this.arrayOfParam[i].poste != null)
				param.poste = this.arrayOfParam[i].poste;
			
			if (this.arrayOfParam[i].resu != null) {
				param.typ = this.arrayOfParam[i].resu.type;
				param.resu = this.arrayOfParam[i].resu.value;
			}
				
			params.push(param);
		}
		
		return {
			json: {
				spgm: this.spgmId,
				result: {
					errp: this.errp || "",
					errm: this.errm || "",
					errn: this.errn || 0,
					errt: this.errt || 0,
					errl: this.errl || 0
				},
				debugInfos: {
					exceptiondetail: this.arrayOfResu[0] ? this.arrayOfResu[0].value || '' : '',
					totalduration: this.arrayOfResu[1] ? this.arrayOfResu[1].value || 0 : 0,
					execduration: this.arrayOfResu[2] ? this.arrayOfResu[2].value || 0 : 0,
					trace: this.arrayOfResu[3] ? this.arrayOfResu[3].value || '' : ''
				},
				"return": result,
				params: params
			}
		}

	},

});
