// syracuse-x3/lib/convergence/model/CvgReport.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var CvgObservable = require('syracuse-x3/lib/convergence/model/CvgObservable').CvgObservable;

var CvgReportSubmitted = exports.CvgReportSubmitted = helpers.defineClass(function(client, properties, error) {
	CvgObservable.call(this, client);
	
	this.host = properties.host;
	this.trackId = properties.trackId;
	this.title = properties.title;
	this.rank = properties.rank;
	this.error = error;
	
	this.kind = "submitted";
	this.setJsonType('REPORT');
}, CvgObservable, {

	toJSON: function() {
		var res = {
	        host: this.host,
	        trackId: this.trackId,

	        title: this.title,
	        rank: this.rank,
	        mime: "application/pdf",
	        facet: "$viewer"
		}
		
		if (this.error) {
	        res.$diagnoses = [{
	          severity: "error",
	          message: this.error.message,
	          stackTrace: this.error.stack
	        }];
		}else {
	        res.$diagnoses = [{
	          severity: "info",
	          message: "Request considered"
	        }];
		}
		return res;
	},
	
});

var CvgReportPending = exports.CvgReportPending = helpers.defineClass(function(client) {
	CvgObservable.call(this, client);
	this.kind = "pending";
	this.setJsonType('REPORT');
}, CvgObservable, {

	toJSON: function() {

	},
	
});

var CvgReportDone = exports.CvgReportDone = helpers.defineClass(function(client) {
	CvgObservable.call(this, client);
	this.kind = "done";
	this.setJsonType('REPORT');
}, CvgObservable, {

	toJSON: function() {
		
	},
	
});
