// syracuse-x3/lib/convergence/model/CvgScreen.js
"use strict";
var helpers = require('syracuse-core').helpers;
var CvgEntity = require('syracuse-x3/lib/convergence/model/CvgEntity').CvgEntity;
var CvgListOfDataSrcs = require('syracuse-x3/lib/convergence/model/CvgListOfDataSrcs').CvgListOfDataSrcs;

var CvgScreen = exports.CvgScreen = helpers.defineClass(function(requesterWindow, idxbaseOne, aScreenId, aListOfEntity) {
	CvgEntity.call(this, requesterWindow && requesterWindow.client, aScreenId, idxbaseOne, 'SCREEN');
	this.requesterWindow = requesterWindow;
	this.listOfEntity = aListOfEntity !== undefined ? aListOfEntity : new CvgListOfDataSrcs(requesterWindow);
	// En dernier
	this.setJsonType('ENTITY');
}, CvgEntity, {
	getAdxId: function() {
		if (this.requesterWindow == null) {
			return null;
		}
		return "" + this.requesterWindow.getAdxId() + ',' + this.entityIdxAlpha;
	},
	getJsonKey: function() {
		return this.getObservableId().substring(1, 2);
	},
	isScreenButtons: function() {
		return this.entityIndexBaseOne === CvgScreen.SCREEN_BUTTONS_NUM_ZERO_BASED + 1;
	},
	isScreenLeftList: function() {
		return this.entityIndexBaseOne === CvgScreen.SCREEN_LEFTLIST_NUM_ZERO_BASED + 1;
	},
	isScreenTabbed: function() {
		return this.entityIndexBaseOne < CvgScreen.SCREEN_BUTTONS_NUM_ZERO_BASED + 1;
	},
	isScreenTabTitle: function() {
		return this.entityIndexBaseOne === CvgScreen.SCREEN_TABTITLES_NUM_ZERO_BASED + 1;
	},
	pushDataSrc: function(aId, aDataSrc) {
		this.listOfEntity.pushEntity(aId, aDataSrc);
	},
	toJSON: function() {
		var object = {};
		// Modif Sreen
		if (this.hasModifTitle()) {
			object.tit = this.title;
			if (this.titleSty != null) {
				object.stytit = this.titleSty;
			}
		}
		if (this.hasModifStyle()) {
			object.sty = this.style;
		}
		return object;
	},
	traverseAll: function(aTraverser) {
		return this.listOfEntity.traverseAll(aTraverser);
	},
	traverseChilds: function(aTraverser) {
		return this.listOfEntity.traverseChilds(aTraverser);
	},
	traverseSettable: function(aStructIst, aSetter) {
		return this.listOfEntity.traverseSettable(aStructIst, aSetter);
	}
});

CvgScreen.screenZeroBaseIdxToWhatScreen = function(aIdx) {
	if (aIdx === CvgScreen.SCREEN_LEFTLIST_NUM_ZERO_BASED) {
		return "33|SCREEN_LEFTLIST(34,AH)";
	} else if (aIdx === CvgScreen.SCREEN_BUTTONS_NUM_ZERO_BASED) {
		return "32|SCREEN_BUTTONS(33,AG)";
	} else if (aIdx === CvgScreen.SCREEN_TABTITLES_NUM_ZERO_BASED) {
		return "34|SCREEN_TABTITLES(35,AI)";
	} else {
		return aIdx.toString();
	}
};
CvgScreen.SCREEN_BUTTONS_NUM_ZERO_BASED = 32; // index base "one" = 33
CvgScreen.SCREEN_LEFTLIST_NUM_ZERO_BASED = 33; // index base "one" =
CvgScreen.SCREEN_TABTITLES_NUM_ZERO_BASED = 34; // index base "one" =
