// syracuse-x3/lib/convergence/model/CvgStatusBar.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

var CvgStatusIcon = exports.CvgStatusIcon = helpers.defineClass(function(aImgId, aActionId, aVisible) {
	this.imgId = aImgId;
	this.actionId = aActionId;
	this.visible = aVisible;
}, null, {
	toJSON: function() {
		// this.actionId = 0 pour les icone supprimés
		var ico = [];
		ico.push(this.visible ? this.actionId : 0);
		ico.push(this.imgId);
		return ico;
	},
	update: function(aActionId, aVisible) {
		this.actionId = aActionId;
		this.visible = aVisible;
	}
});

var CvgStatusBar = exports.CvgStatusBar = helpers.defineClass(function() {
	this.iconsList = {};
	this.text = [];

}, null, {
	addIconDef: function(aNumtext, aText, aImgId, aActionId) {
		if (aImgId === 0) {
			this.setText(aNumtext, aText);
		} else {
			var wVisible = aImgId > 0;
			if (!wVisible) {
				aImgId = aImgId * -1;
			}
			var key = aImgId;
			var wStatusIcon = this.iconsList[key];
			if (wStatusIcon == null) {
				this.iconsList[key] = new CvgStatusIcon(aImgId, aActionId, wVisible);
			} else {
				wStatusIcon.update(aActionId, wVisible);
			}
		}
	},
	setText: function(aNumText, aText) {
		if (aNumText < 1 || aNumText > 2) {
			aNumText = 1; // protection
		}
		if (aText != null && aText.length > 0) {
			this.text[aNumText - 1] = aText;
		} else {
			this.text[aNumText - 1] = null;
		}
	},
	toJSON: function() {
		var statusbar = {};
		console.log(this.text[0] == null);
		if (this.text.length > 0) {
			// do not use this.text.map because map skips entries in sparse arrays
			statusbar.t = [];
			for (var i = 0; i < this.text.length; i++) statusbar.t[i] = this.text[i] || "";
		}
		if (this.iconsList.length > 0) statusbar.iconsList = this.iconsList;
		return statusbar;
	}
});
