// syracuse-x3/lib/convergence/model/CvgVariableDate.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var CvgVariable = require('syracuse-x3/lib/convergence/model/CvgVariable').CvgVariable;
var CvgResu = require('syracuse-x3/lib/convergence/types/CvgResu').CvgResu;
var CvgDate = require('syracuse-x3/lib/convergence/types/CvgDate').CvgDate;

function checkFormat(srcLine,fmt){
	if (srcLine && srcLine.dataSrc.isLeftList()) return srcLine.client.formater.defaultfDateFmt();	
	return fmt;
}

var CvgVariableDate = exports.CvgVariableDate = helpers.defineClass(function(srcLine, entityIdx, name, aID, aTdo, fmt, state) {
	CvgVariable.call(this, srcLine, entityIdx, name, aID, aTdo, checkFormat(srcLine,fmt), state);
}, CvgVariable, {
	clone: function() {
		var wVariable = new CvgVariableDate(this.srcLine, this.entityIndexBaseOne, this.name, this.id, this.dataType, this.format, this.state);
		wVariable.mustBeFormatted = this.mustBeFormatted;
		this.cloneValues(wVariable);
		return wVariable;
	},
	formatX3ToDisplay: function(aData) {
		if (!this.mustBeFormatted)  return aData.value;
		if (aData.value == null || aData.value.length==0 || aData.isDateNull())  return "";
		if (aData.value.length !== 8) {
			this.client.addLogicalErrorMess(this, "callFormater", this.getAdxId(), this.fmtExcepMsg('ERROR_FORMAT_EXCEPTION', 'FORMAT_RTOS', aData.value, "NOT AN ADONIX_DATE YYYYMMDD"));
			return aData.value + " ???";
		}
		return this.callFormater(aData.value);
	},
	formatEditToDisplay: function(aValue) {
		return aValue === "" ? "" : CvgVariable.prototype.formatEditToDisplay.call(this, aValue);
	},
	formatX3ToEdit: function() {
		if (this.serverValue != null && this.serverValue.value =="") return "";
		return CvgVariable.prototype.formatX3ToEdit.call(this);
	},
	setServerResu: function(resu) {
		if (resu != null && CvgDate.DATE_ZERO === resu.value)  resu = null;
		CvgVariable.prototype.setServerResu.call(this,resu);
	},
	unFormatData: function(aValue) {
		if (!this.mustBeFormatted) return aValue;
		if (aValue == null) return null;
		if (aValue.length === 0) return aValue;
		aValue = this.callUnFormater(aValue);
		if (CvgResu.DATE_ZERO === aValue)  aValue = "";
		return aValue;
	}
});
