// syracuse-x3/lib/convergence/model/CvgVariableBlob.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var CvgVariable = require('syracuse-x3/lib/convergence/model/CvgVariable').CvgVariable;

var CvgVariableBlob = exports.CvgVariableBlob = helpers.defineClass(function(srcLine, aEntityIndexBaseOne, aVarName, aID, aTdo, aSize, aFormat, aEntityState, aUrl) {
	CvgVariable.call(this, srcLine, aEntityIndexBaseOne, aVarName, aID, aTdo, aFormat, aEntityState);
	this.indirection = null;
	this.size = aSize;
	// TODO - Non utilisé - this.url = aUrl;
}, CvgVariable, {

	// Pas de format
	acceptEditFormat: function() {
		return false;
	},
	clone: function() {
		return null;
	},
	controlBlobSize: function(aBlob) {
		var wMaxSize = 1024 * Math.pow(2, this.size) - 4;
		if (aBlob.length > wMaxSize) {
			throw new CvgException(this.client, null, this.client.getMessage('ERROR_UPLOAD_BLOB', aBlob.length.toString(), this.id, wMaxSize.toString(), this.size.toString()), true);
		}
	},
	getDisplayValue: function() {
		return "/Fusion/Blob/TODO"; // buildFullUrl();
	},
	hasValueBlob: function() {
		return this.serverValue != null && this.serverValue.hasBlobValue();
	},
	setBrowserUnformattedValue: function(aBuffer) {
		this.setBrowserUnformattedValue(new CvgResu(CvgResu.X3TYP_BLOB, aBuffer));
	},
	/* TODO overload
	setBrowserUnformattedValue: function(aBrowserValue) {
		controlBlobSize(aBrowserValue.geBlobValue());
		CvgVariable.prototype.setBrowserUnformattedValue.call(this,aBrowserValue);
	},
	*/
	setIndirection: function(i) {
		this.indirection = i;
	}
});
