// syracuse-x3/lib/convergence/model/CvgVariableCombo.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var CvgVariableLocalMenu = require('syracuse-x3/lib/convergence/model/CvgVariableLocalMenu').CvgVariableLocalMenu;

var CvgVariableCombo = exports.CvgVariableCombo = helpers.defineClass(function(srcLine, aEntityIndexBaseOne, aVarName, aID, aTdo, aFormat, aEntityState, aLocalMenu, window) {
	CvgVariableLocalMenu.call(this, srcLine, aEntityIndexBaseOne, aVarName, aID, aTdo, aFormat, aEntityState, aLocalMenu, window);
	this.disableList = null;
	this.disableModif = false;
}, CvgVariableLocalMenu, {
	clone: function(line) {
		var wVariable = new CvgVariableCombo(line, this.entityIndexBaseOne, this.name, this.id, this.dataType, this.format, this.state, this.localMenu, this.win);
		wVariable.mustBeFormatted = this.mustBeFormatted;
		this.cloneValues(wVariable);
		return wVariable;
	},
	hasDchModifications: function() {
		return this.disableModif;
	},
	hasGuiModModifications: function() {
		return CvgVariableLocalMenu.prototype.hasGuiModModifications.call(this) || this.disableModif;
	},
	razModification: function() {
		CvgVariableLocalMenu.prototype.razModification.call(this);
		this.disableModif = false;
	},
	setServerDisabledChoice: function(choice) {
		if (choice === 0) {
			this.disableList = [];
			this.disableList.push('');
			this.disableModif = true;
		} else {
			if (this.disableList == null) {
				this.disableList = [];
			}
			if (this.disableList.indexOf(choice) === -1) {
				this.disableList.push(choice);
				this.disableModif = true;
			}
		}
		if (this.disableModif && this.srcLine != null) {
			this.srcLine.comboDchModifSet();
		}
		this.forceChange();
	},
	// arr != null pour les grilles
	toDchJSON: function(arr) {
		if (this.disableModif) {
			if (this.disableList == null || this.disableList.length === 0) {
				if (arr != null) {
					arr.push(null);
				}
				return null;
			} else {
				var disStr = this.disableList.join(',');
				if (arr == null) {
					return disStr;
				} else {
					arr.push(disStr);
					return arr;
				}
			}
		}
		return null;
	},
	toJSON: function() {
		var object = CvgVariableLocalMenu.prototype.toJSON.call(this);
		var dch = this.toDchJSON(null);
		if (dch != null) {
			object.dch = dch;
		}
		return object;
	}
});
