// syracuse-x3/lib/convergence/model/CvgVariableLocalMenu.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var CvgVariable = require('syracuse-x3/lib/convergence/model/CvgVariable').CvgVariable;

var CvgVariableLocalMenu = exports.CvgVariableLocalMenu = helpers.defineClass(function(srcLine, aEntityIndexBaseOne, aVarName, aID, aTdo, aFormat, aEntityState, aLocalMenu) {
	CvgVariable.call(this, srcLine, aEntityIndexBaseOne, aVarName, aID, aTdo, aFormat, aEntityState);
	this.localMenu = aLocalMenu;
}, CvgVariable, {
	acceptEditFormat: function() {
		return false;
	},
	formatEditToDisplay: function(aValue) {
		return aValue;
	},
	getDisplayValue: function() {
		var val = this.serverValue;
		if (val == null) return "";
		if (this.localMenu == null)  throw new Error("no local menu available.");
		if (val.isNum()) return val.value > 0 ? this.localMenu.getLib(val.value) || "" : "";
		return val.value || "";
	},
	getValueForJson: function() {
		return this.getServerValInt() === -999 ? this.getServerValStr() : this.getServerValInt();
	}
});
