// syracuse-tablet/html/js/wc/controls/panels/panelBase.js
"use strict";

var utils = require('syracuse-tablet/html/js/wc/helpers/utils');
var Base = require('syracuse-tablet/html/js/wc/controls/ctrlBase').CtrlBase;

/**
 * 
 */
exports.PanelBase = utils.defineClass(function PanelBase(controller, type, article, opts) {
	Base.call(this, controller, type, article, opts);
}, Base, {
	destroy: function() {
		Base.prototype.destroy.call(this);

		// Do not call destroy since we did not create the adapter but just reference to it
		this.actionAdapter = null;
	},

	/**
	 * This must be called by the creator if this component (usually it's the page)
	 * The component will use the adapter instance to get a list of items to be rendered
	 */
	setActionAdapter: function(actionAdapter) {
		this.actionAdapter = actionAdapter;
	},

	//
	// START: Rendering
	// 
	buildHtml: function() {
		this.createRootElement(["s-m-panel", "s-m-control", "s-m-ctrl"]);
	},
	//
	// END: Rendering
	//

	createId: function(typeName) {
		return utils.readableuid(typeName);
	},
	/**
	 * Built with the rank of the layout in generation process to be reliable
	 */
	getUnitTestId: function() {
		if (this.options.index == null) return;
		return utils.unitTestId(this.typeName.toLowerCase(), this.options.index);
	},
	getGestureMgr: function() {
		return null;
	}
});
