// syracuse-tablet/html/js/wc/controls/panels/panelsHtml.js
"use strict";

/* 
 * 
 * Html templates used by more that one panel
 * 
 */

var eventListener = require('syracuse-tablet/html/js/wc/app/eventListener');

var _iconClassMap = {
	"home": "fa fa-home",
	"pin": "fa fa-star-o",
	"unpin": "fa fa-star",
	"openactionspanel": "fa fa-ellipsis-h",
	"switchContext": "fa fa-user",
	"mail": "fa fa-envelope",
	"settings": "fa fa-gear",
	"about": "fa fa-info-circle",
	"development": "fa fa-wrench",
	"logout": "fa fa-sign-out",
	"drafts": "fa fa-building-o",
	"designPage": "fa fa-thumb-tack",

	"accept": "fa fa-check",
	"cancel": "fa fa-reply",


	"$save": "fa fa-save",
	"$edit": "fa fa-edit",
	"$delete": "fa fa-trash-o",
	"$create": "fa fa-plus-square",
	"$query": "fa fa-list",
};

var _html = {
	footer: {
		defaultItem: '<a href="#"\
			draggable="false" \
			data-menuitem-id="{{uniqueId}}" \
			class="s-m-link"\
			{{{attrs}}}>\
			<i class="{{iconClass}}"></i>\
			<span>{{title}}</span>\
		</a>'
	},
	// same as footer except no title is rendered
	header: {
		defaultItem: '<a href="#"\
			draggable="false" \
			data-menuitem-id="{{uniqueId}}" \
			class="s-m-link"\
			{{{attrs}}}>\
			<i class="{{iconClass}}"></i>\
		</a>'
	},
	actionPanel: {
		defaultItem: '<li data-menuitem-id="{{uniqueId}}"\
			class="s-m-link">\
			<a href="#"\
				draggable="false" \
				data-nevent-ctrl-click=""\
				{{{attrs}}}>\
				<i class="{{iconClass}}"></i>\
				<span>{{title}}</span>\
			</a>\
		</li>'
	},
	globalPanel: {
		defaultItem: '<li data-menuitem-id="{{uniqueId}}">\
			<a href="#"\
				draggable="false" \
				class="s-m-link"\
				data-nevent-ctrl-click=""\
				{{{attrs}}}>\
				<i class="{{iconClass}}"></i>\
				<span class="isLabel">{{title}}</span>\
			</a>\
		</li>',
		draftsItem: '<li data-menuitem-id="{{uniqueId}}">\
			<a href="#" \
				draggable="false" \
				data-nevent-ctrl-click=""\
				class="s-m-link {{#if disabled}} disabled"{{/if}}" {{#if disabled}} data-disabled{{/if}}>\
				<i class="{{iconClass}}"></i>\
				<span class="isLabel">{{title}}<span class="badge" style="float: right;">{{count}}</span></span>\
			</a>\
		</li>'
	}
};

var _templates = {};

function _getHtml(panel, name, ctx) {
	var tmpl = _templates[panel];
	if (!tmpl) {
		tmpl = _templates[panel] = {};
	}
	tmpl = tmpl[name];
	if (!tmpl) {
		tmpl = _templates[panel][name] = Handlebars.compile(_html[panel][name]);
	}
	return tmpl(ctx);
};

/**
 * 
 */
exports.getItemHtml = function(panelName, item) {
	var html;
	var tpl;
	var attrs;
	var controllerId = item["controller-id"];

	if (item.action) {
		if (item.action === "pageDrafts") {
			tpl = "draftsItem";
			attrs = eventListener.getDataActionAttrs(controllerId, item.action, item.params, true);
			item.count = item.params;
			item.disabled = item.count < 1;
		} else {
			tpl = "defaultItem";
			attrs = eventListener.getDataActionAttrs(controllerId, item.action, item.params, true);
		}
		item.uniqueId = item.action; // Will be used to find item in the DOM later to apply dynamic meta data
	} else if (item.$link) {
		tpl = "defaultItem";
		attrs = eventListener.getSdataLinkAttrs(controllerId, item.$link, null, true);
		item.uniqueId = item.$link; // Will be used to find item in the DOM later to apply dynamic meta data
	} else if (item.$action) {
		tpl = "defaultItem";
		attrs = eventListener.getSdataActionAttrs(controllerId, item.$action, true);
		item.uniqueId = item.$action; // Will be used to find item in the DOM later to apply dynamic meta data
	}

	if (tpl) {
		item.iconClass = _iconClassMap[item.icon];
		item.attrs = attrs;
		html = _getHtml(panelName, tpl, item);
	}

	return html;
}
