// @sage/syracuse-ui/lib/fusion/fusionBar.js
"use strict";
var helpers = require('@sage/syracuse-core').helpers;
var SideBar = require("syracuse-ui/lib/page/tools/sideBar").SideBar;

function FusionBar() {}

exports.FusionBar = helpers.defineClass(FusionBar, SideBar, {
	load: function(page, $fusionBar) {
		this.minWidth = 300;
		this.pagePreferenceKey = "convergenceBar";
		this.options = {
			leftSide: true,
			$skin: "s-fusion-bar"
		};
		if (syra_site.isTabletDevice) {
			this.options.isAutoModeDisabled = true;
			this.options.$isFloating = true;
		}

		this.slot = syra_dom.div("s-fusion-bar-slot");
		this.body = syra_dom.div("s-fusion-bar-body");
		this.$fusionBar = $fusionBar;
		page.domItem.insertBefore(this.slot, page.domItem.firstChild);
		SideBar.prototype.load.call(this, page);
		var count = this.$fusionBar.$items.length;
		this.blocks = [];
		for (var ii = 0; ii < count; ii++) {
			this.blocks.push(this.createBlock(this.$fusionBar.$items[ii], ii, count));
		}
		if (!this.openedBlock && count > 0) {
			this.collapseBlock(this.blocks[0], true, true);
		}
	},

	createBlock: function($item, index, count) {
		var block = {
			boxParent: this,
			index: index,
			id: helpers.uuid.generate(),
			$item: $item
		};
		var fusionSess = (syra_fusion.syraUtil.getFusionController(this.page)).getSession();
		$item.$XID = $item.$XID || $item.$bind;
		$item.$title = $item.$title || this.page.$prototype.$properties[$item.$bind].$title;

		var $title = $item.$title || "";
		if ($title.length > 0 && $title[1] == "@") {
			$title = syra_expression.render(this.page, $title);
		}

		block.header = syra_dom.header("s-fusion-block-header", this.body);
		block.header.id = block.id;
		var btn = syra_button.add({
			parent: this,
			slot: block.header,
			text: $title,
			css: "s-fusion-bar-block-title",
			click: function(event, btn) {
				var bar = this.parent;
				for (var ii = 0, jj = bar.blocks.length; ii < jj; ii++) {
					if (bar.blocks[ii].title == btn) {
						bar.collapseBlock(bar.blocks[ii], true);
						break;
					}
				}
			}
		});
		block.title = btn.link;
		block.title.setAttribute("href", "#");
		block.body = syra_dom.div("s-fusion-block-body", this.body);
		syra_dom.hide(block.body, true);
		return block;
	},
	collapseBlock: function(block, show, onDrawBar) {
		var self = this;
		self.page.externalAdapter.onBoxToggle({
			box: block,
			open: show,
			onDrawParent: onDrawBar,
			isFirstTime: !block.leftList,
			doEvent: function() {
				if (show && self.openedBlock && (self.openedBlock != block)) {
					self.collapseBlock(self.openedBlock, false);
				}
				self.openedBlock = block;
				syra_dom.hide(block.body, !show);
				syra_dom.toggleClass(block.header, "s-open", show);
				if (show) {
					if (!block.leftList) {
						block.$item.$isTitleHidden = true;
						if (self.page.$prototype.$properties[block.$item.$bind].$type == "application/x-array") {
							block.$item.$isNavigationList = true;
						}
						syra_dom.empty(block.body);
						block.leftList = self.page.addItem(block.body, block.$item);
						block.title.syraKeyTarget = block.leftList.id;
					} else {
						if (block.leftList) {
							block.leftList.disableDisplayValidation = false;
							block.leftList.filler.validateDisplay(block.leftList);
						}
					}
					if (!onDrawBar) {
						block.leftList && block.leftList.resizeItem(true);
					}
				} else {
					if (block.leftList) {
						block.leftList.disableDisplayValidation = true;
					}
				}
			}
		});
	},
	focus: function(select, bindList) {
		var block, ii = 0,
			jj = this.blocks ? this.blocks.length : 0,
			ret = false;
		for (ii = 0; ii < jj; ii++) {
			block = this.blocks[ii];
			if (block.$item && block.$item.$bind === bindList) {
				break;
			}
		}
		if (ii < jj) {
			ret = true;
			block.title.focus();
		}
		return ret;
	},
	dispose: function() {
		if (this.blocks) {
			for (var ii = 0, jj = this.blocks.length; ii < jj; ii++) {
				delete this.blocks[ii].leftList;
			}
			delete this.blocks;
		}
		this.$fusionBar = null;
		SideBar.prototype.dispose.call(this);
	}
});
