// syracuse-trace/lib/contract.js
"use strict";

exports.contract = {
	application: "syracuse",
	contract: "trace",
	entities: {
		traceConfig: require('./entities/traceConfig').entity,
		traceConfigModule: require('./entities/traceConfigModule').entity,
		traceSessionConfig: require('./entities/traceSessionConfig').entity,
		traceSetting: require('./entities/traceSetting').entity,
		traceExportSelect: require('./entities/traceExportSelect').entity,
		trace: require('syracuse-trace/lib/entities/trace').entity
	},
	representations: {
		//trace: require('syracuse-trace/lib/representations/trace').representation
	},
	dbMeta: {
		initScript: [],
		updateScript: [{
			script: __dirname + '/updateScripts/updateScript',
			version: 1
		}]
	}
};

exports.extendContract = function(contract) {
	function copy(src, dst) {
		Object.keys(src).forEach(function(name) {
			dst[name] = src[name];
		});
	}
	copy(exports.contract.entities, contract.entities);
	copy(exports.contract.representations, contract.representations);
	contract.dbMeta.updateScript.push(exports.contract.dbMeta.updateScript[0]);
};
