// @sage/syracuse-react/src/test/field/diagnoseState.tsx
"use strict";
import * as React from 'react';
import { ChoiceField } from '../../field/poc/choice/choiceField';

export class DiagnoseState extends React.Component<{
    onChange: (meta: any) => void
}, {
        severity: string
    }>{
    constructor(props: any, context: any) {
        super(props, context);
        this.state = {
            severity: "no"
        };
    }
    diagnoses: any = {
        error: {
            $severity: "error",
            $message: "Error test  blalfoz"
        },
        warning: {
            $severity: "warning",
            $message: "Warning test"
        },
        info: {
            $severity: "info",
            $message: "Info test"
        }
    };
    onChange = (value: string) => {
        this.props.onChange({
            $diagnoses: value== "no"?null: [this.diagnoses[value]]
        });
        this.setState({
            severity: value
        });
        return true;
    }
    render() {
        return <ChoiceField meta={{
            $value: {
                $enum: [{
                    $title: "error",
                    $value: "error"
                }, {
                    $title: "warning",
                    $value: "warning"
                }, {
                    $title: "info",
                    $value: "info"
                }, {
                    $title: "no",
                    $value: "no"
                }]
            }
        }}
            title="$diagnoses"
            editable={true}
            value={this.state.severity}
            onChange={this.onChange} />
    }
}
