// syracuse-soap/lib/common/soapTracer.js
"use strict";

var config = require('config');
var helpers = require("syracuse-core/lib/helpers");

var SoapTracer = helpers.defineClass(function(name) {
	//	console.log("TRACER="+name +": "+(config && config.soap && config.soap[name].tracer));
	this.tracer = config && config.soap && config.soap[name] && config.soap[name].tracer != null ? config.soap[name].tracer : false;
}, null, {
	traceSection: function(title, content) {
		this.trace("\n");
		this.trace("##################################################");
		this.trace("#    " + title);
		this.trace("##################################################");
		if (content != null && content !== "") this.trace(content);
	},
	traceSubSection: function(title, content) {
		this.trace("\n");
		this.trace("--------------------------------------------------");
		this.trace("|    " + title);
		this.trace("--------------------------------------------------");

		if (content != null && content !== "") this.trace(content);
	},
	trace: function(str) {
		if (this.tracer === true) console.log(" " + str);
	},
	warning: function(str) {
		if (this.tracer === true) this.trace("## WARNING ## : " + str);
	}
});

exports.create = function(name) {
	return new SoapTracer(name);
};
