// syracuse-ui/lib/article/layouts/layoutProvider.js
"use strict";

function _extractItem($items, $layout){
    $layout.$items.forEach(function($item){
        if ($item.$layoutType) {
            _extractItem($items, $item);
        }
        else {
            $items.push($item);
        }
    });
}

function _buildAutoBlock($prototype, $parent, $blocks){
    var $row;
    if ($blocks.length > 1) {
        $parent.$layout.$items = [];
        $blocks.forEach(function($block){
            var $children = [];
            _extractItem($children, $block.$layout);
            
            
            var $fields = [];
            _extractItem($fields, $block.$layout);
            if ($fields.length > 0) {
                var addAuto = true;
                if ($fields.length == 1) {
                    var $field = $fields[0];
                    if ($field.$bind && $prototype.$properties[$field.$bind]) {
                        if ($prototype.$properties[$field.$bind].$type == "application/x-array") {
                            $row = null;
                            addAuto = false;
                        }
                    }
                }
                if (addAuto) {
                    _buildAutoFields($prototype, $block, $fields);
                }
                
                
            }
            if (!$row) {
                $row = {
                    $layoutType: "row",
                    $items: []
                };
                $parent.$layout.$items.push($row);
            }
            $row.$items.push($block);
            if ($row.$items.length == 3) {
                $row = null;
            }
        });
    }
    else 
        if ($blocks.length == 1) {
            var $fields = [];
            _extractItem($fields, $blocks[0].$layout);
            _buildAutoFields($prototype, $blocks[0], $fields);
        }
}

function _buildAutoFields($prototype, $parent, $children){
    $parent.$layout.$items = [{
        $layoutType: "columns",
        $layoutSubType: "50,50",
        $items: [{
            $layoutType: "stack",
            $items: $children
        }, {
            $layoutType: "stack",
            $items: []
        }]
    }];
}

exports.buildAutoLayout = function($prototype, $sourceArticle){
    var $newArticle = {
        $layout: {
            $items: []
        }
    };
    var $rootItems = [];
    _extractItem($rootItems, $sourceArticle.$layout);
    if ($rootItems[0].$category != "section") {
        $newArticle.$layout.$items.push({
            $category: "section",
            $title: "Section",
            $layout: {
                $items: $rootItems
            }
        });
    }
    else {
        if ($rootItems.length < 3) {
            $newArticle.$layout.$items = $rootItems;
        }
        else {
            $newArticle.$layout.$items.push({
                $layoutType: "stack",
                $items: [$rootItems[0]]
            });
            var $tabs = {
                $layoutType: "tabs",
                $items: []
            };
            $newArticle.$layout.$items.push($tabs);
            $newArticle.$layout.$items.push({
                $layoutType: "stack",
                $items: [$rootItems[$rootItems.length - 1]]
            });
            for (var ii = 1, jj = $rootItems.length - 1; ii < jj; ii++) {
                $tabs.$items.push($rootItems[ii]);
            }
        }
    }
    $rootItems.forEach(function($parent){
        var $children = [];
        _extractItem($children, $parent.$layout);
        if ($children.length > 0) {
            if ($children[0].$category == "block") {
                _buildAutoBlock($prototype, $parent, $children);
            }
            else {
                _buildAutoFields($prototype, $parent, $children);
            }
        }
    });
    return $newArticle;
}
