// syracuse-ui/lib/article/layouts/layout.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function Layout(){
}

exports.Layout = helpers.defineClass(Layout, null, {
    create: function($layout, boxParent){
        this.boxParent = boxParent;
        this.page = this.boxParent.page;
        this.$authoringLevel = "layout";
        var article = this.boxParent.articleParent || this.boxParent;
        this.id = this.boxParent.id + "-" + (++article._childLayoutOffset);
        article.layouts[this.id] = this;
        this.isLayout = true;
        this.$item = this.$layout = $layout;
        if (!$layout.$items) {
            $layout.$items = [];
        }
        if (!this.$layout.$layoutType) {
            this.$layout.$layoutType = "stack";
        }
        else {
            if (this.$layout.$layoutType == "side") {
                this.$layout.$layoutType = "row";
                this.$layout.$fitToContent = true;
            }
            if (this.$layout.$layoutType == "columns") {
                this.$layout.$layoutType = "row";
            }
            if (this.$layout.$layoutSubType) {
                this.$layout.$widths = this.$layout.$layoutSubType;
            }
        }
    },
    getTitle: function(){
        return "";
    },
    applyDesignMetaData: function(metaData, onAuthoring){
        if (onAuthoring !== false) {
            if (metaData.$fieldsIsTopLabelAlignment !== undefined) {
                this.$item.$fieldsIsTopLabelAlignment = metaData.$fieldsIsTopLabelAlignment;
            }
            if (metaData.$fieldsIsRightTextLabelAlignment !== undefined) {
                this.$item.$fieldsIsRightTextLabelAlignment = metaData.$fieldsIsRightTextLabelAlignment;
            }
            if (metaData.$field !== undefined) {
                this.$item.$field = this.$item.$field || {};
                var $props = Object.keys(metaData.$field);
                for (var ii = 0, jj = $props.length; ii < jj; ii++) {
                    var $prop = $props[ii];
                    this.$item.$field[$prop] = metaData.$field[$prop];
                }
            }
        }
    },
    loadChildItem: function(childItem, $childItem, index){
        if (childItem) {
            if (childItem && childItem.layoutParent) {
                childItem.layoutParent.extractItem(childItem);
            }
            return this.appendChildItem(childItem, index, false);
        }
        return this.createChildItem($childItem, null, index);
    },
    appendChildItem: function(item, index, load){
        item.layoutParent = this;
        item.boxParent = this.boxParent;
        if (item.isSpaceBox) {
            this.hasSpaceBox = true;
        }
        var append;
        if (!item.layoutSlot) {
            item.layoutSlot = document.createElement("div");
        }
        item.isTabLayout = this.$layout.$layoutType == "tabs";
        if (index === undefined) {
            if (this.isItemRegisterDisable == false) {
                this.$layout.$items.push(item.$item);
            }
            this.items.push(item);
            this._item.appendChild(item.layoutSlot);
        }
        else {
            if (this.isItemRegisterDisable == false) {
                this.$layout.$items.splice(index, 0, item.$item);
            }
            this.items.splice(index, 0, item);
            var sibling = this.items[index + 1];
            if (sibling) {
                sibling.layoutSlot.parentNode.insertBefore(item.layoutSlot, sibling.layoutSlot);
            }
            else {
                if (item.isTabLayout) {
                    append = true;
                }
                this._item.appendChild(item.layoutSlot);
            }
        }
        
        
        if (this.$layout.$layoutType == "row") {
            if (this.$layout.$valign) {
                item.layoutSlot.style.verticalAlign = this.$layout.$valign;
            }
            else {
                if (item.$field) {
                    //  item.layoutSlot.style.verticalAlign = "bottom";
                }
            }
        }
        item.ensureLayoutMode();
        if (load) {
            if (item.isLayout) {
                item.render();
            }
            else {
                item.loadBox();
            }
        }
        if (item.isTabLayout) {
            if (sibling && sibling._tabTitle && item._tabTitle) {
                sibling._tabTitle.parentNode.insertBefore(item._tabTitle, sibling._tabTitle);
            }
            else {
                if (append) {
                    this._tabs.appendChild(item._tabTitle);
                }
            }
        }
        return item;
    },
    createChildItem: function($item, item, index){
        item = this.page.createNewItem($item, this.boxParent, item);
        if (item) {
            this.appendChildItem(item, index, true);
        }
        return item;
    },
    extractItem: function(item){
        var index = this.getItemIndex(item, true);
        if (index >= 0) {
            this.items.splice(index, 1);
        }
        index = this.getItemIndex(item.$item);
        if (index >= 0) {
            this.$layout.$items.splice(index, 1);
        }
        if (item.isTabLayout) {
            if (item._tabTitle) {
                document.site.removeDomChild(item._tabTitle);
            }
        }
        if (item.hasSpaceBox && item.layoutParent) {
            item.layoutParent.hasSpaceBox = false;
        }
        item.layoutParent = null; //mettre a null test
        return item;
    },
    grabItems: function(sourceLayout, index, excludeItem){
        var lastMoved;
        for (var ii = sourceLayout.items.length - 1; ii >= 0; ii--) {
            if (!excludeItem || (sourceLayout.items[ii] != excludeItem)) {
                lastMoved = this.loadChildItem(sourceLayout.items[ii], null, index);
            }
        }
        return lastMoved;
    },
    removeItem: function(item, removeSlot){
        this.extractItem(item);
        if (removeSlot && item.layoutSlot) {
            document.site.removeDomChild(item.layoutSlot);
        }
        this.boxParent.getArticle().removeItem(item, true);
    },
    getItemIndex: function($item, byItems){
        var list = byItems ? this.items : this.$layout.$items;
        for (var ii = 0; ii < list.length; ii++) {
            if (list[ii] == $item) {
                return ii;
            }
        }
        return -1;
    },
    addSpaceBox: function(index){
        if (this.boxParent && this.boxParent.isMenuGroup) {
            return null;
        }
        return this.createChildItem({
            $category: "space"
        }, null, index);
    },
    removeSpaceBox: function(){
        var spaces = [];
        for (var ii = 0, jj = this.items.length; ii < jj; ii++) {
            if (this.items[ii].isSpaceBox) {
                spaces.push(this.items[ii]);
            }
        }
        if (spaces.length == this.items.length) {
            spaces.shift();
        }
        for (var ii = 0, jj = spaces.length; ii < jj; ii++) {
            this.removeItem(spaces[ii], true);
        }
    },
    ensureLayoutSlot: function(item){
        if (item.isField) {
            item.layoutSlot.className = item.$skin + "-slot s-field-layout-slot s-slot-" + this.$layout.$layoutType;
        }
        else {
            if (item.$skin) {
                item.layoutSlot.className = item.$skin + "-layout-slot s-slot-" + this.$layout.$layoutType;
            }
            else {
                item.layoutSlot.className = "s-slot-" + this.$layout.$layoutType;
            }
            
        }
    },
    ensureLayoutMode: function(){
        if (this.layoutParent) {
            this.layoutParent.ensureLayoutSlot(this);
        }
        if (!this._item) {
            this._item = document.createElement("div");
            this._item.setAttribute("data-s-layout", this.id);
            this.layoutSlot.appendChild(this._item);
            this.$$item = $(this._item);
        }
        if (this.layoutParent) { // && this.$layout.$layoutType == "stack") {
            this._item.className = "s-layout-" + this.$layout.$layoutType;
        }
        else {
            this._item.className = this.boxParent.$skin + "-layout s-layout-" + this.$layout.$layoutType;
        }
        if (this.page.authoringPage) {
            this.page.authoringPage.toggleItemAuthoring(this, true);
        }
    },
    render: function(){
        this.ensureLayoutMode();
        this.items = [];
        this.isItemRegisterDisable = true;
        for (var ii = 0, jj = this.$layout.$items.length; ii < jj; ii++) {
            this.createChildItem(this.$layout.$items[ii]);
        }
        this.isItemRegisterDisable = false;
        this.ensureEndUpdateLayout();
        if (this.page.authoringPage) {
            this.page.authoringPage.toggleItemAuthoring(this, true);
        }
    },
    appendTab: function(box){
        if (!this._tabs) {
            this._tabs = document.createElement("nav");
            this._tabs.className = box.$skin + "-tabs-nav";
            this.layoutSlot.className = box.$skin + "-layout-tabs-slot s-slot-" + (this.layoutParent ? (this.layoutParent.$layout.$layoutType) : "stack");
            $(this._item).prepend(this._tabs);
        }
        this._tabs.appendChild(box._tabTitle);
    },
    ensureEndUpdateLayout: function(){
        switch (this.$layout.$layoutType) {
            case "row":
                var $widths;
                if (this.$layout.$widths) {
                    $widths = this.$layout.$widths.split(",");
                    if ($widths.length > this.items.length) {
                        for (var ii = this.items.length, jj = $widths.length; ii < jj; ii++) {
                            this.addSpaceBox();
                        }
                    }
                }
                else {
                    var width = 100 / this.items.length;
                    $widths = [];
                    for (var ii = 0, jj = this.items.length; ii < jj; ii++) {
                        $widths.push(Math.floor(width));
                    }
                    this.$layout.$widths = $widths.join(",");
                }
                if (!this.$layout.$fitToContent) {
                    this._item.style.width = "100%";
                }
                for (var ii = 0, jj = $widths.length; ii < jj; ii++) {
                    this.items[ii].layoutSlot.style.width = $widths[ii] + "%";
                }
                this.tabOpened = null;
                break;
            case "tabs":
                if (this.items.length > 0) {
                    var opened = [];
                    for (var ii = 0, jj = this.items.length; ii < jj; ii++) {
                        if (this.items[ii].$item.$opened) {
                            opened.push(this.items[ii])
                        }
                    }
                    if (opened.length > 0) {
                        if (opened.length > 1) {
                            for (var ii = 0, jj = opened.length; ii < jj; ii++) {
                                opened.openBox(false);
                            }
                            this.tabOpened = this.items[0];
                        }
                        else {
                            this.tabOpened = opened[0];
                        }
                    }
                    else {
                        this.tabOpened = this.items[0];
                    }
                    if (this.tabOpened) {
                        this.tabOpened.openBox(true, true);
                    }
                }
                else {
                    this.tabOpened = null;
                }
                this.$layout.$widths = null;
                this._item.style.width = "";
                break;
            case "stack":
                if (this.items.length == 0) {
                    this.addSpaceBox();
                }
                else {
                    if (this.hasSpaceBox && this.items.length > 1) {
                        this.removeSpaceBox();
                    }
                }
                for (var ii = 0, jj = this.items.length; ii < jj; ii++) {
                    this.items[ii].layoutSlot.style.width = "";
                }
                this.tabOpened = null;
                this.$layout.$widths = null;
                this._item.style.width = "";
                break;
        }
    },
    dispose: function(){
        this.$$item = this._item = this._tabs = this.items = this.layoutParent = this.boxParent = this.page = this.layoutSlot = this.$item = this.$layout = this.tabOpened = null;
    }
});
