// syracuse-ui/lib/article/layouts/spaceBox.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Box = require("syracuse-ui/lib/article/box").Box;

function SpaceBox() {}

exports.SpaceBox = helpers.defineClass(SpaceBox, Box, {
	drawBox: function() {
		this.isSpaceBox = true;
		this.$skin = "s-aw-empty";
		this.$field = {};
		this.$authoringLevel = "field";
		this.$$item = this.$$body = $(this._item = this._header = this._body = document.createElement("div"));
		this.layoutSlot.appendChild(this._item);
		this.ensureLayoutMode();
		this.loaded = true;
	},
	getTitle: function() {
		return "";
	},
	setTitle: function() {},
	ensureLayoutMode: function() {
		if (this._item) {
			this._item.className = this.$skin;
			if (this.layoutParent) {
				this.layoutParent.ensureLayoutSlot(this);
			}
			if (this.page.authoringPage) {
				this.page.authoringPage.toggleItemAuthoring(this, true);
			}
		}
	},
	openBox: function(open) {},
	onMouseEnter: function() {

	},
	dispose: function() {
		this.$field = this.deleteLink = null;
		Box.prototype.dispose.call(this);
	}
});
