// syracuse-ui/lib/article/layouts/tabsLayout.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Layout = require('./layout').Layout;

function TabsLayout(){
}

exports.TabsLayout = helpers.defineClass(TabsLayout, Layout, {
    _wrapInSlot: function(item){
        if (item.$$header) {
            item.applyCollapsable(false);
            item.$$header.remove();
        }
        item.$isTabLayout = true;
        item.layoutParent = this;
        item.$$container = this.$$container;
        var $$prevItem = item.$$item;
        if ($$prevItem) {
            item.appendHeader();
            item.$$body.removeClass(item.$skin + "-body");
            this.wrapTabBody(item);
            item.$$item.toggle(this.tabOpened == item);
            this._setOpenSkin(item, this.tabOpened == item);
            $$prevItem.remove();
        }
    },
    _ensureTabs: function(){
        if (!this._$$tabs) {
            this._$$tabs = $("<nav/>").prependTo(this.$$container);
        }
    },
    _initializeTabs: function(box){
        var self = this;
        self._isTabsInitialized = true;
        self._ensureTabs();
        self._$$tabs.addClass(box.$skin + "-tabs-nav").delegate("a." + (self.$tabSkin = box.$skin + "-tab"), "click", function(event, data){
            if (!document.site.DDAuthoring) {
                var tabIdx = $(this).index(), tabItem = self.items[tabIdx];
                var isTabSelected = self.tabOpened == tabItem;
                var open = !tabItem.$item.$opened;
                var isFirstTime = !tabItem.loaded;
                tabItem.getPage().externalAdapter.onBoxToggle({
                    nativeEvent: event,
                    nativeEvenData: data,
                    box: tabItem,
                    isTabSelected: isTabSelected,
                    open: open,
                    isFirstTime: isFirstTime,
                    tabIdx: tabIdx,
                    doEvent: function(){
                        if (!isTabSelected) {
                            tabItem.openBox(open);
                        }
                        document.site.resize();
                    }
                });
                return tabItem.authoringNode ? true : false; //authoringNode => true hooked by layout
            }
        }).delegate("a." + box.$skin + "-tab" + "-open", "click", function(event){
            var tabIdx = $(this).index(), tabItem = self.items[tabIdx];
            tabItem.getPage().externalAdapter.onBoxClick({
                box: tabItem,
                event: event,
                tabIdx: tabIdx
            });
            return false;
        });
    },
    appendChildTab: function(box){
        box.domTitle = document.createElement("a");
        box.domTitle.className = box.$skinTab = box.$skin + "-tab";
        box.domTitle.setAttribute("href", "#");
        box.$$header = $(box.domTitle);
        if (!this._isTabsInitialized) {
            this._initializeTabs(box);
        }
        box.$$header.appendTo(this._$$tabs);
        if (box.authoringNode) {
            box.authoringNode.setDesignableItem(box, true);
        }
    },
    openTab: function(item, open){
        if (open) {
            if (this.tabOpened !== undefined) {
                this.tabOpened.openBox(false);
            }
            this.tabOpened = item;
            item.$item.$opened = true;
        }
        else {
            delete item.$item.$opened;
        }
        this._setOpenSkin(item, open);
    },
    _setOpenSkin: function(item, open){
        item.$$header[0].className = item.$$header[0].className.replace(item.$skinTab + (open ? "" : "-open"), item.$skinTab + (open ? "-open" : ""));
    },
    
    wrapTabBody: function(box){
        box.$$item = box.$$body.appendTo(box.$$container).addClass(box.$skin + "-tab-body");
    },
    
    _ensureOneTabOpened: function(){
        if (this.items.length > 0) {
            (this.tabOpened = this.items[0]).openBox(true, true);
        }
    },
    appendNewItem: function(options){
        Layout.prototype.appendNewItem.call(this, options);
        if (options.action) {
            options.newItem.$$header[options.action](options.targetItem.$$header);
        }
        if (!this.tabOpened) {
            this._ensureOneTabOpened();
        }
    },
    renderChildItem: function(item){
        if (item.$item.$opened) {
            this.tabOpened = item;
        }
        item.$isTabLayout = true;
        Layout.prototype.renderChildItem.call(this, item);
    },
    render: function(refresh){
        Layout.prototype.render.call(this, refresh);
        this._ensureTabs();
        this._ensureOneTabOpened();
    },
    
    dispose: function(){
        //disposeOk
        if (this._$$tabs) {
            this._$$tabs.undelegate().remove();
            delete this._$$tabs;
            delete this.tabOpened;
        }
        Layout.prototype.dispose.call(this);
    }
});
