// hdb/lib-cov/protocol/index.js
/* automatically generated by JSCoverage - do not edit */
if (typeof _$jscoverage === 'undefined') _$jscoverage = {};
if (! _$jscoverage['protocol/index.js']) {
  _$jscoverage['protocol/index.js'] = [];
  _$jscoverage['protocol/index.js'][14] = 0;
  _$jscoverage['protocol/index.js'][16] = 0;
  _$jscoverage['protocol/index.js'][17] = 0;
  _$jscoverage['protocol/index.js'][18] = 0;
  _$jscoverage['protocol/index.js'][19] = 0;
  _$jscoverage['protocol/index.js'][20] = 0;
  _$jscoverage['protocol/index.js'][21] = 0;
  _$jscoverage['protocol/index.js'][22] = 0;
  _$jscoverage['protocol/index.js'][23] = 0;
  _$jscoverage['protocol/index.js'][24] = 0;
  _$jscoverage['protocol/index.js'][25] = 0;
  _$jscoverage['protocol/index.js'][26] = 0;
  _$jscoverage['protocol/index.js'][27] = 0;
  _$jscoverage['protocol/index.js'][28] = 0;
  _$jscoverage['protocol/index.js'][29] = 0;
  _$jscoverage['protocol/index.js'][30] = 0;
  _$jscoverage['protocol/index.js'][31] = 0;
  _$jscoverage['protocol/index.js'][32] = 0;
}
_$jscoverage['protocol/index.js'][14]++;
"use strict";
_$jscoverage['protocol/index.js'][16]++;
exports.Connection = require("./Connection");
_$jscoverage['protocol/index.js'][17]++;
exports.Parser = require("./Parser");
_$jscoverage['protocol/index.js'][18]++;
exports.Result = require("./Result");
_$jscoverage['protocol/index.js'][19]++;
exports.ResultSet = require("./ResultSet");
_$jscoverage['protocol/index.js'][20]++;
exports.Statement = require("./Statement");
_$jscoverage['protocol/index.js'][21]++;
exports.Stringifier = require("./Stringifier");
_$jscoverage['protocol/index.js'][22]++;
exports.Transaction = require("./Transaction");
_$jscoverage['protocol/index.js'][23]++;
exports.Reader = require("./Reader");
_$jscoverage['protocol/index.js'][24]++;
exports.Writer = require("./Writer");
_$jscoverage['protocol/index.js'][25]++;
exports.ExecuteTask = require("./ExecuteTask");
_$jscoverage['protocol/index.js'][26]++;
exports.Lob = require("./Lob");
_$jscoverage['protocol/index.js'][27]++;
exports.auth = require("./auth");
_$jscoverage['protocol/index.js'][28]++;
exports.common = require("./common");
_$jscoverage['protocol/index.js'][29]++;
exports.data = require("./data");
_$jscoverage['protocol/index.js'][30]++;
exports.part = require("./part");
_$jscoverage['protocol/index.js'][31]++;
exports.reply = require("./reply");
_$jscoverage['protocol/index.js'][32]++;
exports.request = require("./request");
_$jscoverage['protocol/index.js'].source = ["// Copyright 2013 SAP AG.","//","// Licensed under the Apache License, Version 2.0 (the \"License\");","// you may not use this file except in compliance with the License.","// You may obtain a copy of the License at","//","// http: //www.apache.org/licenses/LICENSE-2.0","//","// Unless required by applicable law or agreed to in writing,","// software distributed under the License is distributed on an","// \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,","// either express or implied. See the License for the specific","// language governing permissions and limitations under the License.","'use strict';","","exports.Connection = require('./Connection');","exports.Parser = require('./Parser');","exports.Result = require('./Result');","exports.ResultSet = require('./ResultSet');","exports.Statement = require('./Statement');","exports.Stringifier = require('./Stringifier');","exports.Transaction = require('./Transaction');","exports.Reader = require('./Reader');","exports.Writer = require('./Writer');","exports.ExecuteTask = require('./ExecuteTask');","exports.Lob = require('./Lob');","exports.auth = require('./auth');","exports.common = require('./common');","exports.data = require('./data');","exports.part = require('./part');","exports.reply = require('./reply');","exports.request = require('./request');"];
