// syracuse-ui/lib/office/excel/desktopTopPanel.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Article = require("syracuse-ui/lib/common/article/article").Article;

var _localize = {
    $about: "About",
    $drafts: "Drafts",
    $historic: "Historic",
    $sessions: "Sessions",
    $bookmarks: "Bookmarks",
    $dashboards: "Dashboard",
    $bookmarkLinkTitle: "Add this page?",
    $bookmarkLinkDesc: "It will appear in your bookmark",
    $authoringLinkTitle: "Design the current page?",
    $authoringLinkDesc: "Launch the authoring panel"
};

function DesktopTopPanel(){

}

helpers.defineClass(DesktopTopPanel, Article, {
    appendOpener: function($$container){
        var self = this;
        self.$$opener = $("<a id='s-site-top-pn-opener-link'/>").appendTo($$container).bind("click", function(event){
            if (self._topPanelTimeout) {
                window.clearTimeout(self._topPanelTimeout);
                delete self._topPanelTimeout;
            }
            
            self._topPanelTimeout = setTimeout(function(){
                if (!self.$$container) {
                    self.$$item = self.$$container = $("<div id='s-site-top-pn'/>").insertAfter(document.site._$$header);
                    self.loadBox();
                }
                self._togglePanel(self.$$container.css("height") == "0px");
            }, 10);
            return false;
        });
        
    },
    _togglePanel: function(show, animate){
        this.$$opener.toggleClass("s-open", show);
        if (show) {
            var $$item = document.site.userProfile ? document.site.userProfile.$$item : null;
            if ($$item) {
                if ($$item.parent().length == 0) {
                    this.$$topUserProfile.empty().append($$item);
                }
            }
            else {
                this.$$topUserProfile.empty();
            }
        }
        if (animate !== false) {
            this.$$container.animate({
                height: show ? ((this.$$body.outerHeight() + this.$$body.offset().top) + "px") : "0px"
            }, 500);
        }
        else {
            this.$$container.css("height", show ? ((this.$$body.outerHeight() + this.$$body.offset().top) + "px") : "0px");
        }
    },
    renderTopLink: function(options){
        var $$link = $("<a class='s-top-link'/>");
        $$link.append($("<div class='s-top-link-icon'/>").addClass(options.$icon));
        $$link.append($("<div class='s-top-link-title'/>").text(options.$title).append($("<div class='s-top-link-desc'/>").text(options.$desc)));
        $$link.bind("click", options.$click).appendTo(this.$$topLinks);
    },
    renderTopBar: function(){
        var self = this;
        self.$$topBar = $("<div id='s-site-top-pn-top-bar'/>").appendTo(self.$$body);
        self.$$topUserProfile = $("<div id='s-site-top-pn-user-profile'/>").appendTo(self.$$topBar);
        
        self.$$topLinks = $("<div id='s-site-user-top-pn-links'/>").appendTo(self.$$topBar);
    },
    renderNavigationPanel: function(navItem){
        navItem.$$cell = $("<div/>").addClass("s-site-top-pn-nav-bar-cell").addClass(navItem.$css).appendTo(this.$$navigationBar);
        navItem.$$title = $("<div/>").text(navItem.$title).addClass("s-site-top-pn-nav-title").appendTo(navItem.$$cell);
        navItem.links = document.itemFactory.load($("<div/>").addClass("s-site-top-pn-nav-item-body").appendTo(navItem.$$cell), {
            $category: "links",
            $isBindDisabled: navItem.$isBindDisabled || false,
            $links: navItem.$links,
            $skin: "s-site-top-pn-nav-bar-menus"
        }, this);
        navItem.$$cell.bind("mouseleave mouseenter", function(event){
            navItem.links.$$item.find("a").toggleClass("s-site-top-pn-nav-bar-cell-over", event.type == "mouseenter");
            return false;
        });
    },
    renderAboutBar: function(){
        this.$$aboutBar = $("<div id='s-site-top-pn-about-bar'/>").appendTo(this.$$body);
        document.itemFactory.load(this.$$aboutBar, {
            $category: "block",
            $title: _localize.$about,
            $skin: "s-site-top-pn-about-bar",
            $layout: {
                $items: [{
                    $category: "links",
                    $isBindDisabled: true,
                    $links: this.$prototype.$navigation.$about,
                    $skin: "s-site-top-pn-about-menus"
                }]
            }
        }, this);
    },
    drawBox: function(){
        this.$$body = $("<div id='s-site-top-pn-body'/>").appendTo(this.$$item);
        this.renderTopBar();
        this.renderAboutBar();
    },
    dispose: function(){
        var self = this;
        if (self.navItems) {
            Object.keys(self.navItems).forEach(function(navItem){
                if (navItem.$$cell) {
                    navItem.$$cell.unbind();
                }
            });
        }
    }
});

exports.appendPanel = function($$container){
    var panel = document.site.desktopTopPanel = new DesktopTopPanel();
    panel.$prototype = document.site.$prototype;
    document.itemFactory.initializeItem(panel, {});
    panel.appendOpener($$container);
};
