// syracuse-ui/lib/office/excel/excelSite.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Site = require('syracuse-ui/lib/site/site').Site;
var store = require('syracuse-ui/lib/site/tools/store');


exports.load = function($item, $prototype) {
	window.syra_store = store;
	var widgetsLibrary = require('syracuse-ui/lib/site/widgetsLibrary');
	widgetsLibrary.pageCategories.excelreport = require("./excelReportPage").ExcelReportPage;
	$item.$isSecurityTitleVisible = false;
	var site = new Site();
	site.$isSaveSite = $item.$isSaveSite;
	site._onAfterLogon = function() {
		syra_site.history.start();
		try {
			if (this.$isSaveSite) {
				var docUrl = this.excelInterface.getDocumentUrl();
				if (docUrl && docUrl !== "") {
					if (/syracuse\/collaboration\/syracuse/.test(docUrl)) {
						this.excelInterface.publishDocument();
					} else {
						this.excelInterface.publishDocumentX3(docUrl);
					}
				}
			} else {
				this.excelInterface.onLogin();
			}
		} catch (e) {}
		this.layoutSlot.style.display = "";
		this.resize();
	};
	site._renderHeader = function() {

	};
	var controllerClass;
	if (site.$isSaveSite) {
		controllerClass = require("syracuse-ui/lib/office/excel/controller").Controller;
	} else {
		controllerClass = require("syracuse-ui/lib/controller/controller").Controller;
	}
	site.loadBox({
		isOffice: true,
		widgetsLibrary: widgetsLibrary,
		controllerClass: controllerClass,
		isUserEndpointChangeDisabled: true,
		$item: $item,
		$prototype: $prototype
	});

	var excelInterface = require('msoffice/lib/excel/excelInterface').ExcelInterface;
	site.excelInterface = new excelInterface();
	site.$syracuseMainPageUrl = $item.$syracuseMainPageUrl;
	return site;
};

//Called from word when additional data has to be fetched (eg. images)
//ONLY WORKS IN IE. no problem here, since this is only called from IE within word!
window.readBinaryURLContentIE = function(url, returnData) {
	try {
		var xhr = new XMLHttpRequest();
		var data = {};
		xhr.open('GET', url, false);

		xhr.setRequestHeader("X-Content-Type-Override", "text/plain; charset=x-user-defined");
		xhr.onreadystatechange = function(e) {};
		xhr.send();
		if (xhr.status == 200) {
			// this.responseText is broken in IE with binary data!
			// This is a c# byte array!
			returnData.setData(xhr.responseBody);
		} else
		if (xhr.status == 1223) {
			returnData.setErrorText("NOTFOUND");
		} else
		if (xhr.status == 404) {
			returnData.setErrorText("NOTFOUND");
		} else {
			returnData.setErrorText(xhr.response + "(" + xhr.status + ")");
		}
	} catch (e) {
		returnData.setErrorText(e);
	}
};
