// syracuse-ui/lib/office/excel/excelWorksheetActions.js
"use strict";
exports.getPrototype = function(reprName, facetName, onSuccess, onError) {
	var result = {};
	if (facetName == "$query") {
		result.$prototype = {
			$properties: {},
			$links: {
				refreshAll: {
					$title: "Refresh all",
					$url: "excel://excelDatasources/$service/refreshAll?representation=excelDatasource.$query",
					$method: "POST"
				}
			}
		};
		onSuccess(result);
	}
};

exports.dispatcher = {
	GET: function(excelQuery, id, data, onSuccess, onError) {},
	PUT: function(excelQuery, id, data, onSuccess, onError) {
		var refreshAll = data && data.$actions && data.$actions.refreshAll;
		if (refreshAll && refreshAll.$isRequested) {
			syra_site.excelDocument.refreshAllDatasources();
			refreshAll.$isRequested = false;
		}
		onSuccess(null);
	}
};

exports.executeOperation = function(name, id, method, data, onSuccess, onError) {
	// saveDocument - POST
	if (syra_site.excelDocument.documentUrl)
		syra_site.excelDocument.publishDocument("", null, null, function(data, response) {
			// show diagnoses as server doesn't send diagnoses in this case
			syra_site.showDiagnoses({
				$diagnoses: [{
					severity: "info",
					message: "The document has been saved"
				}]
			});
		});
	else {
		external.ShowSettingsForm();
	}
};
