// syracuse-ui/lib/office/excel/excelWorksheetActions.js
"use strict";
exports.getPrototype = function(reprName, facetName, onSuccess, onError) {
	var result = {};
	if (facetName == "$query") {
		result.$prototype = {
			$properties: {},
			$links: {
				refreshAll: {
					$title: "Refresh all",
					$url: "excel://excelDatasources/$service/refreshAll?representation=excelDatasource.$query",
					$method: "POST"
				}
			}
		};
		onSuccess(result);
	}
};

var _operationMap = {
	saveDocument: {
		POST: function(id, data, onSuccess, onError) {
			if (syra_site.excelDocument.documentUrl)
				syra_site.excelDocument.publishDocument("", null, null, function(data, response) {
					// show diagnoses as server doesn't send diagnoses in this case
					syra_site.showDiagnoses({
						$diagnoses: [{
							severity: "info",
							message: "The document has been saved"
						}]
					});
				});
			else
				external.ShowSettingsForm();
		}
	}
};
exports.dispatcher = {
	GET: function(excelQuery, id, data, onSuccess, onError) {},
	PUT: function(excelQuery, id, data, onSuccess, onError) {
		var refreshAll = data && data.$actions && data.$actions.refreshAll;
		if (refreshAll && refreshAll.$isRequested) {
			syra_site.excelDocument.refreshAllDatasources();
			refreshAll.$isRequested = false;
		}
		onSuccess(null);
	}
};

exports.executeOperation = function(name, resourceId, method, data, onSuccess, onError) {
	var op = _operationMap[name] && _operationMap[name][method];
	if (!op)
		throw new Error("Operation unknown " + name + " for entity excelDatasources");
	op(resourceId, data, onSuccess, onError);
};
