// syracuse-ui/lib/office/excel/officeConfigSite.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var siteHelper = require("syracuse-ui/lib/common/site/siteHelper");
var Popup = require('syracuse-ui/lib/desktop/site/popup').Popup;
var Dialog = require('syracuse-ui/lib/desktop/site/dialog').Dialog;
var Article = require("syracuse-ui/lib/common/article/article").Article;
var Controller = require("./controller").Controller;
var ExcelDocument = require('msoffice/lib/excel/excelDocument').ExcelDocument;
var desktopTopPanel = require("./desktopTopPanel");
var dispatcher = require('msoffice/lib/excel/dispatcher');

var _$siteOptions = {
    widgetsLibrary: require('syracuse-ui/lib/desktop/site/widgetsLibrary'),
    messageBoxClass: require('syracuse-ui/lib/desktop/site/messageBox').MessageBox,
    diagnoseViewerClass: require('syracuse-ui/lib/desktop/site/diagnoseViewer').DesktopDiagnoses
};

function OfficeSite(){
}

exports.OfficeSite = helpers.defineClass(OfficeSite, Article, {
    notifyRequestSurvey: function(pendingRequests){
        if (!this.$$requestSurvey) {
            this.$$requestSurvey = $("<div id='s-site-request'/>").appendTo(this.$$container), this.$$requestSurveyCount = $("<label id='s-site-request-count'/>").appendTo(this.$$requestSurvey);
        }
        pendingRequests = pendingRequests || 0;
        this.$$requestSurveyCount.text(pendingRequests);
        if (pendingRequests > 0) {
            if (this.$$requestSurvey.css("right") != "0px") {
                this.$$requestSurvey.css("right", "0px");
            }
        }
        else {
            this.$$requestSurvey.css("right", "-80px");
        }
    },
    openDialog: function(options){
        var dialog = new Dialog();
        dialog.open(options);
        return dialog;
    },
    onMainPageChange: function($itemPage){
        this.$$body[0].style.display = "none";
        $itemPage.boxParent = null;
        $itemPage.$$container = this.$$body;
        this._releaseMainPage();
        if (this._diagnoseViewer) {
            this._diagnoseViewer._emptyViewer();
        }
        this.mainPage = document.itemFactory.loadPage($itemPage);
        this.mainPage.isMainPage = true;
        //
        this.updateDocumentTitle();
        this.$$body[0].style.display = "";
        return this.mainPage;
    },
    drawBox: function(){
        var self = this;
        document.authoringSite = {}; //
        self._renderHeader();
        self.$$body = $("<div id='s-site-body'/>").appendTo(self.$$container);
        $(window).bind("resize", function(evt){
            self.resize();
        });
        document.site.logon(function(){
            document.controller.startNavigation();
            self.$$container.show();
            self.resize();
            try {
                external.onLogon();
            } 
            catch (e) {
                // allows Firefox debug
            }
        });
    },
    onNotifyRecordChange: function(value, binding){
        if (binding == "search") {
            this.dataset[binding] = value;
            document.controller.executeMenu(this.menuItems["$searchLink"][0]);
        }
    },
    updateDocumentTitle: function(){
        var title = this.mainPage.getTitle();
        if (title) {
            var record = this.userProfile.dataset;
            if (record) {
                if (record.selectedRole) {
                    title += " (" + record.selectedRole.description + ")";
                }
                if (record.selectedEndpoint) {
                    title += " (" + record.selectedEndpoint.description + ")";
                }
            }
            else {
                this.userProfile.$isTitleUpdateRequested = true;
            }
            this.mainPage.drawUserProfile(record);
            document.title = title;
        }
    },
    createPopup: function(options){
        return (new Popup()).load(options);
    },
    resize: function(){
        if (this._$$header && this.$$body) {
            var $$header = this._$$header;
            var newHeight = $$header.outerHeight(true);
            this.$$body[0].style.height = (document.site.$$container.height() - newHeight) + "px";
        }
        var self = this;
        if (this.dialogAdapter) {
            self.dialogAdapter.onDocumentResize();
        }
    },
    
    _renderHeader: function(){
        var self = this;
        self._$$header = $("<header id='s-site-header'>/");
        var $$top = $("<div id='s-site-header-top'>");
        document.site.$$popupContainer = $("<div id='s-site-popup'/>").appendTo(document.site.$$container);
        
        var $$topLeft = $("<div id='s-site-header-top-left'/>").appendTo($$top);
        
        
        $$topLeft.append("<div id='s-site-sage'/>").append($("<div id='s-site-module'/>").text(self.$item.$module));
        
        desktopTopPanel.appendPanel($$top);
        
        var $$topRight = $("<div id='s-site-header-top-right'/>").appendTo($$top);
        
        document.itemFactory.load($("<div id='s-site-home'/>").appendTo($$topRight), {
            $bind: "$home",
            $category: "link",
            $noText: true,
            $skin: "s-site-home-link"
        }, this);
        
        document.itemFactory.load($("<div id='s-site-help'/>").appendTo($$topRight), {
            $bind: "$help",
            $category: "link",
            $noText: true,
            $skin: "s-site-help-link"
        }, this);
        
        self._$$header.append($$top).appendTo(self.$$container);
        self.appendDiagnoseViewer(self.$$container, self);
    },
    _renderSearch: function($$container){
    }
});

exports.load = function($item, $prototype){
    _$siteOptions.siteClass = OfficeSite;
    _$siteOptions.userProfileClass = require("./userProfile").UserProfile;
    _$siteOptions.controllerClass = Controller;
    _$siteOptions.$item = $item;
    _$siteOptions.$prototype = $prototype;
    _$siteOptions.widgetsLibrary.pageCategories.worksheet = require("./excelPage").ExcelPage;
    //
    var site = siteHelper.loadSite(_$siteOptions);
    site.$syracuseMainPageUrl = $item.$syracuseMainPageUrl;
    //
    site.agents = {
        excelDatasources: {
            saveDatasource: function(article, menu, record){
                site.excelDocument.addDatasource(record);
            }
        }
    }
    // must be done after all initialization as it connects and load page
    site.excelDocument = new ExcelDocument();
    dispatcher.on("refreshAll", function(){
        site.excelDocument.refreshAllDatasources();
    });
    //
    return site;
};
