// syracuse-ui/lib/office/excel/officeSite.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var siteHelper = require("syracuse-ui/lib/common/site/siteHelper");
var Popup = require('syracuse-ui/lib/desktop/site/popup').Popup;
var Article = require("syracuse-ui/lib/common/article/article").Article;
var Controller = require("./controller").Controller;
var DialogDecorator = require('syracuse-ui/lib/desktop/site/dialogDecorator').DialogDecorator;
var ExcelDocument = require('msoffice/lib/excel/excelDocument').ExcelDocument;
var dispatcher = require('msoffice/lib/excel/dispatcher');

var _$siteOptions = {
    widgetsLibrary: require('syracuse-ui/lib/desktop/site/widgetsLibrary'),
    messageBoxClass: require('syracuse-ui/lib/desktop/site/messageBox').MessageBox,
    diagnoseViewerClass: require('syracuse-ui/lib/desktop/site/diagnoseViewer').DesktopDiagnoses
};

function OfficeSite(){
}

exports.OfficeSite = helpers.defineClass(OfficeSite, Article, {
    notifyRequestSurvey: function(pendingRequests){
        if (!this.$$requestSurvey) {
            this.$$requestSurvey = $("<div id='s-site-request'/>").appendTo(this.$$container), this.$$requestSurveyCount = $("<label id='s-site-request-count'/>").appendTo(this.$$requestSurvey);
        }
        pendingRequests = pendingRequests || 0;
        this.$$requestSurveyCount.text(pendingRequests);
        if (pendingRequests > 0) {
            if (this.$$requestSurvey.css("right") != "0px") {
                this.$$requestSurvey.css("right", "0px");
            }
        }
        else {
            this.$$requestSurvey.css("right", "-80px");
        }
    },
    
    onMainPageChange: function($itemPage){
        this.$$body[0].style.display = "none";
        $itemPage.boxParent = null;
        $itemPage.$$container = this.$$body;
        this._releaseMainPage();
        if (this._diagnoseViewer) {
            this._diagnoseViewer._emptyViewer();
        }
        this.mainPage = document.itemFactory.loadPage($itemPage);
        this.mainPage.isMainPage = true;
        this.updateDocumentTitle();
        this.$$body[0].style.display = "";
        return this.mainPage;
    },
    drawBox: function(){
        var self = this;
        document.authoringSite = {}; //
//        self._renderHeader();
        self.$$userProfileHost = $("<div id='s-user-profile'/>").appendTo(self.$$container);
        self.$$body = $("<div id='s-site-body'/>").appendTo(self.$$container);
        $(window).bind("resize", function(evt){
            self.resize();
        });
        document.site.logon(function(){
            document.controller.startNavigation();
//            self.$$userProfileHost.show();
            self.$$container.show();
            self.resize();
            //
            try {
                external.onLogon();
            } catch(e) {
                // allows Firefox debug
            }
        });
    },
    onNotifyRecordChange: function(value, binding){
        if (binding == "search") {
            this.dataset[binding] = value;
            document.controller.executeMenu(this.menusFacade.items["$searchLink"][0]);
        }
    },
    updateDocumentTitle: function(){
        var title = this.mainPage.getTitle();
        if (title) {
            var record = this.userProfile.dataset;
            this.mainPage.drawUserProfile(record);
            
            document.title = title;
        }
    },
    applyDialogDecorator: function(view){
        return (new DialogDecorator()).wrap(view);
    },
    removeDialog: function(){
        document.controller.disposeObject(this.dialogAdapter);
        delete this.dialogAdapter;
    },
    createPopup: function(options){
        return (new Popup()).load(options);
    },
    resize: function(){
        if (this._$$header && this.$$body) {
            var $$header = this._$$header;
            var newHeight = $$header.outerHeight(true);
            this.$$body[0].style.height = (document.site.$$container.height() - newHeight) + "px";
        }
        var self = this;
        if (this.dialogAdapter) {
            self.dialogAdapter.onDocumentResize();
        }
    },   
    _renderHeader: function(){
    },
    _renderSearch: function($$container){
    }
});

exports.load = function($item, $prototype){
    _$siteOptions.siteClass = OfficeSite;
    _$siteOptions.userProfileClass = require("./userProfile").UserProfile;
    _$siteOptions.controllerClass = Controller;
    _$siteOptions.$item = $item;
    _$siteOptions.$prototype = $prototype;
    _$siteOptions.widgetsLibrary.pageCategories.worksheet = require("./excelPage").ExcelPage;
    //
    var site = siteHelper.loadSite(_$siteOptions);
    site.excelDocument = new ExcelDocument();
    dispatcher.on("refreshAll", function() {
        site.excelDocument.refreshAllDatasources();
    });
    dispatcher.on("saveDocument", function() {
        site.excelDocument.publishDocument();
    });
    dispatcher.on("selectionChanged", function() {
        site.excelDocument.selectionChanged();
    });
    site.$syracuseMainPageUrl = $item.$syracuseMainPageUrl;
    return site;
};
