// syracuse-ui/lib/office/excel/officeSiteTopPanel.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Article = require("syracuse-ui/lib/article/article").Article;
var locale = require('syracuse-core/lib/locale');

function OfficeSiteTopPanel(){

}

exports.OfficeSiteTopPanel = helpers.defineClass(OfficeSiteTopPanel, Article, {
    loadBox: function(){
        this.page.initializeNewItem(this, {});
        document.site.setZIndex((this.$$item = this.$$container)[0]);
        Article.prototype.loadBox.call(this);
        var $$item = document.site.userProfile ? document.site.userProfile.$$item : null;
        if ($$item) {
            if ($$item.parent().length == 0) {
                this.$$topUserProfile.empty().append($$item);
            }
        }
        this.$$container.css("height", (this.$$body.outerHeight() + this.$$body.offset().top) + "px");
    },
    _buildItems: function($links){
        var ids = Object.keys($links);
        var $menus = [];
        for (var ii = 0, jj = ids.length; ii < jj; ii++) {
            var id = ids[ii];
            this.$prototype.$links[id] = $links[id];
            $menus.push({
                $bind: id
            });
        }
        return $menus;
    },
    drawBox: function(){
        this.$$body = $("<div id='s-site-top-pn-body'/>").appendTo(this.$$item);
        var $$topBar = $("<div id='s-site-top-pn-top-bar'/>").appendTo(this.$$body);
        this.$$topUserProfile = $("<div id='s-site-top-pn-user-profile'/>").appendTo($$topBar);
        $("<div id='s-site-user-top-pn-links'/>").appendTo($$topBar);
        var localize = locale.resources(module)();
        this.page.loadNewItem(this.$$body, {
            $category: "menus",
            $title: localize.office_panel_about,
            $skin: "s-site-top-pn-about-bar",
            $layout: {
                $items: this._buildItems(this.$prototype.$navigation.$about)
            }
        }, this);
    },
    closePanel: function(){
        var $$item = document.site.userProfile ? document.site.userProfile.$$item : null;
        if ($$item) {
            $$item.detach();
        }
        this.$$container.hide().empty();
        document.controller.disposeObject(this);
    }
});
