// syracuse-ui/lib/office/excel/officeSite.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var siteHelper = require("syracuse-ui/lib/common/site/siteHelper");
var DesktopSite = require('syracuse-ui/lib/desktop/site/desktopSite').DesktopSite;
var Controller = require("./controller").Controller;
var ExcelDocument = require('msoffice/lib/excel/excelDocument').ExcelDocument;
var dispatcher = require('msoffice/lib/excel/dispatcher');

function OfficeSite(){
}

exports.OfficeSite = helpers.defineClass(OfficeSite, DesktopSite, {
    drawBox: function(){
        var self = this;
        self._renderHeader();
        self.$$body = $("<div id='s-site-body'/>").appendTo(self.$$container);
        $(window).bind("resize", function(evt){
            self.resize();
        });
        document.site.logon(function(){
            document.controller.startNavigation();
            //            self.$$userProfileHost.show();
            self.$$container.show();
            self.resize();
            //
            try {
                external.onLogon();
            } 
            catch (e) {
                // allows Firefox debug
            }
        });
    },
    updateDocumentTitle: function(){
        var title = this.mainPage.getTitle();
        if (title) {
            var record = this.userProfile.dataset;
            this.mainPage.drawUserProfile(record);
            
            document.title = title;
        }
    },
    _renderHeader: function(){
        this.$$userProfileHost = $("<div id='s-user-profile'/>").appendTo(this.$$container);
    },
    _renderSearch: function($$container){
    }
});

exports.load = function($item, $prototype){
    var widgetsLibrary = require('syracuse-ui/lib/desktop/site/widgetsLibrary');
    widgetsLibrary.pageCategories.worksheet = require("./excelPage").ExcelPage;
    var site = siteHelper.loadSite({
        siteClass: OfficeSite,
        widgetsLibrary: widgetsLibrary,
        messageBoxClass: require('syracuse-ui/lib/desktop/site/messageBox').MessageBox,
        diagnoseViewerClass: require('syracuse-ui/lib/desktop/site/diagnoseViewer').DesktopDiagnoses,
        userProfileClass: require("./userProfile").UserProfile,
        controllerClass: Controller,
        $item: $item,
        $prototype: $prototype
    });
    site.excelDocument = new ExcelDocument();
    dispatcher.on("refreshAll", function(){
        site.excelDocument.refreshAllDatasources();
    });
    dispatcher.on("saveDocument", function(){
        site.excelDocument.publishDocument();
    });
    dispatcher.on("selectionChanged", function(){
        site.excelDocument.selectionChanged();
    });
    site.$syracuseMainPageUrl = $item.$syracuseMainPageUrl;
    return site;
};
