// syracuse-ui/lib/article/box.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function _toggleTabSiblings(box, show) {
	var tab = box.tabTitle;
	var node = tab.parentNode.firstChild;
	while (node) {
		if (node != tab) {
			node.style.display = show ? "" : "none";
		}
		node = node.nextSibling;
	}
}

function _onMaximizeBoxClick(box) {
	var page = box.page;
	if (page.isLandingPage && syra_site.landingPageMaster) {
		page = syra_site.landingPageMaster;
	}
	var slot = page.body;
	box.isMaximized = !box.isMaximized;
	if (box.isMaximized) {
		box._parentChildren = [];
		var nodes = slot.children;
		for (var ii = 0, jj = nodes.length; ii < jj; ii++) {
			var node = nodes[ii];
			node.parentNode && node.parentNode.removeChild(node);
			box._parentChildren.push(node);
		}
		if (box.tabTitle) {
			slot.appendChild(box.layoutParent._tabs);
			_toggleTabSiblings(box, false);
			box.layoutParent.openTab(box);
		}
		slot.appendChild(box.domItem);
	} else {
		if (box.tabTitle) {
			if (box.layoutParent.domItem.firstChild) {
				box.layoutParent.domItem.firstChild.parentNode.insertBefore(box.layoutParent._tabs, box.layoutParent.domItem.firstChild);
			} else {
				box.layoutParent.domItem.appendChild(box.layoutParent._tabs);
			}
			_toggleTabSiblings(box, true);
		}
		box.layoutSlot.appendChild(box.domItem);
		syra_site.dom.empty(slot);
		for (var ii = 0, jj = box._parentChildren.length; ii < jj; ii++) {
			slot.appendChild(box._parentChildren[ii]);
		}
		delete box._parentChildren;
	}
	var state = box.isMaximized ? "minimize" : "maximize";
	syra_menus.updateButtonIcon(box.btns.items.maximize, syra_local["box_" + state], state);

	box.page.menuBar && box.page.menuBar.onChildPageMaximize(box.isMaximized);
	box.page.fusionBar && box.page.fusionBar.onChildPageMaximize(box.isMaximized);
	syra_site.landingPageMaster && syra_site.landingPageMaster.bar.onChildPageMaximize(box.isMaximized);
	box.onSwitchMaximize && box.onSwitchMaximize();
};

var _onClickHandlers = {
	maximize: "onMaximizeBoxClick",
	design: "onDesignOpenerClick",
	"delete": "onDeleteClick",
	refresh: "onRefreshClick",
	close: "onCloseClick"
};

function _showExpandButton(box, designing) {
	if (box.$item.$isBoxCollapsable && !box.$item.$isTitlePicker) {
		if (!box.expandBtn) {
			var css = box.$skin + "-expand";
			box.expandBtn = syra_menus.addIconButton(syra_local.box_collapse, css, "onOpenerClick", null, "expand");
			box.expandBtn.syraItem = box.id;
			box.header.insertBefore(box.expandBtn, box.header.firstChild);
		}
		box.$opened = box.$item.$opened !== false;
		_setExpandButtonState(box);
	} else {
		box.expandBtn && syra_site.dom.removeChild(box.expandBtn);
	}
	if (designing) {
		if (!box.$item.$isBoxCollapsable && !box.$opened) {
			box.expandBody(true);
		}
	}
}

function _setExpandButtonState(box) {
	var state = box.$opened ? "collapse" : "expand";
	syra_menus.updateButtonIcon(box.expandBtn, syra_local["box_" + state], state);
}

function _showButton(box, btnId, show, syraOnClick, fontIcon) {
	var slotParent = box.tabTitleHead ? box.tabTitleHead : box.header;
	var btns = box.btns;
	if (show) {
		if (!btns) {
			box.btnWidth = box.btnWidth || 20;
			btns = box.btns = {
				items: {},
				slot: document.createElement("div"),
				width: 0
			};
			btns.slot.syraItem = box.id;
			btns.slot.className = "s-section-btns";
			if (slotParent) {
				slotParent.appendChild(btns.slot);
			}
		}
		if (!btns.items[btnId]) {
			var btn = syra_menus.addIconButton(syra_local["box_" + btnId], box.$skin + "-btn", _onClickHandlers[btnId] || syraOnClick, null, fontIcon || btnId);
			btns.items[btnId] = btn;
			btns.width += box.btnWidth;
			btns.slot.style.width = btns.width + "px";
			btns.slot.appendChild(btn);
		}
	} else {
		if (btns && btns.items[btnId]) {
			syra_site.dom.removeChild(btns.items[btnId]);
			delete btns.items[btnId];
			btns.width = Math.max(0, btns.width - box.btnWidth);
			btns.slot.style.width = btns.width + "px";
		}
	}
}

function _onCloseClick(box) {
	box.isMaximized && _onMaximizeBoxClick(box);
	var layoutParent = box.layoutParent;
	layoutParent.removeItem(box, true);
	layoutParent.page.ensurePageVisibility();
}

function _applyButtonMetadata(box, metaData, designing) {
	if (metaData.$isBoxCollapsable !== undefined) {
		box.$item.$isBoxCollapsable = metaData.$isBoxCollapsable;
		_showExpandButton(box, designing);
	}
	if (metaData.$isCloseable !== undefined) {
		_showButton(box, "close", box.$item.$isCloseable = metaData.$isCloseable);
	}
	if (metaData.$isMaximizable !== undefined) {
		_showButton(box, "maximize", box.$item.$isMaximizable = metaData.$isMaximizable);
	}
	box.ensureButtonsVisibility();
}


function _expandSiblingBoxes(box, layout, expand) {
	for (var ii = 0, jj = layout.items.length; ii < jj; ii++) {
		var item = layout.items[ii];
		if (item.isLayout) {
			_expandSiblingBoxes(box, item, expand);
		} else {
			if (item.isSection && item != box) {
				item.expandBody(expand);
			}
		}
	}
}


function _setTitleIcon(box) {
	var icon = document.createElement("div");
	icon.className = box.$skin + "-title-icon";
	box.domTitle.appendChild(icon);
	var $path = syra_site.$item.$iconPath + (box.$item.$titleIcon.$path || "");
	icon.style.backgroundImage = "url('" + $path + box.$item.$titleIcon.$value + ".png')";
	box.titleLabel = document.createElement("div");
	box.titleLabel.className = box.$skin + "-title-text s-icon";
	box.domTitle.appendChild(box.titleLabel);
	box.domTitle.syrainout = box.id;
	if (box.$item.$titleIcon.$mode === "icon") {
		box.isTitleIconOnly = true;
		box.domTitle.style.display = "";
		box.titleLabel.style.display = "none";
	}
}

function _addTabTitle(box) {
	box.tabTitle = document.createElement("a");
	box.tabTitle.syraOnClick = "onOpenerClick";
	if (box.isSection) {
		box.tabTitle.setAttribute("href", "#");
	}
	box.tabTitle.className = box.$skin + "-tab";
	box.tabTitle.syraItem = box.id;
	box.tabTitleHead = document.createElement("div");
	box.tabTitleHead.className = box.$skin + "-tab-head";
	box.tabTitleText = document.createElement("div");
	box.tabTitleText.className = box.$skin + "-tab-text";
	box.tabTitleHead.appendChild(box.tabTitleText);
	box.tabTitle.appendChild(box.tabTitleHead);
	box.setTitle(box.getTitle());
	box.layoutParent.appendTab(box);
	box.header.style.display = "none";
}

function _applyDesignFieldsSettings(box, metaData) {
	if (metaData.$fieldsIsTopLabelAlignment !== undefined ||
		metaData.$fieldsIsRightTextLabelAlignment !== undefined ||
		metaData.$fieldsIsTitleHidden !== undefined) {
		if (metaData.$fieldsIsTopLabelAlignment !== undefined) {
			box.$item.$fieldsIsTopLabelAlignment = metaData.$fieldsIsTopLabelAlignment;
		}
		if (metaData.$fieldsIsRightTextLabelAlignment !== undefined) {
			box.$item.$fieldsIsRightTextLabelAlignment = metaData.$fieldsIsRightTextLabelAlignment;
		}
		if (metaData.$fieldsIsTitleHidden !== undefined) {
			box.$item.$fieldsIsTitleHidden = metaData.$fieldsIsTitleHidden;
		}
		var items = box.layoutContent.getItems();
		for (var ii = 0, jj = items.length; ii < jj; ii++) {
			var item = items[ii];
			if (item.isField) {
				box.applyDesignMetaDataToField(item, metaData);
			} else {
				if (!item.isLayout && item.layoutContent) {
					_applyDesignFieldsSettings(item, metaData);
				}
			}
		}
	}
}

function Box() {}

exports.Box = helpers.defineClass(Box, null, {
	onMaximizeBoxClick: function() {
		_onMaximizeBoxClick(this);
	},
	onCloseClick: function() {
		_onCloseClick(this);
	},
	onOpenerClick: function(event, btn) {
		var self = this;
		syra_site.siteFunctions.onBeforeClick(self, event);
		if (!self.page.onOpenerChildBoxClick || self.page.onOpenerChildBoxClick(self, btn)) {
			if (self.tabTitle) {
				var tabIndex = self.layoutParent.getChildIndex(self);
				var isTabSelected = self.layoutParent.getOpenedTab() == self;
				if (isTabSelected) {
					self.page.externalAdapter.onBoxClick({
						box: self,
						event: event,
						tabIdx: tabIndex
					});
				} else {
					if (!syra_dd.ddAgent) {
						var open = !self.$item.$opened;
						var isFirstTime = !self.loaded;
						self.page.externalAdapter.onBoxToggle({
							nativeEvent: event,
							nativeEvenData: btn.syraOnEventData,
							box: self,
							isTabSelected: isTabSelected,
							open: open,
							isFirstTime: isFirstTime,
							tabIdx: tabIndex,
							doEvent: function() {
								if (!isTabSelected) {
									self.layoutParent.openTab(self);
								}
							}
						});
						delete btn.syraOnEventData;
						return false;
					}
				}
			} else {
				if (self.$item.$isBoxCollapsable) {
					self.expandBody(undefined, event);
					if (self.page && self.page.designer) {
						self.$item.$opened = self.$opened;
						self.page.designer.endArticleUpdate(self, true);
					}
				}
			}
		}
	},
	getDefaultTitle: function() {
		return syra_local["box_" + this.$designLevel + "Title"];
	},
	applyDesignMetaDataToField: function(field, metaData) {
		field.applyDesignMetaData({
			$isTopLabelAlignment: metaData.$fieldsIsTopLabelAlignment,
			$isRightTextLabelAlignment: metaData.$fieldsIsRightTextLabelAlignment,
			$isTitleHidden: metaData.$fieldsIsTitleHidden
		}, true);
	},
	applyDesignMetaData: function(metaData, designing) {
		this.setState(metaData);
		if (metaData.$isTitleHidden !== undefined) {
			this.$item.$isTitleHidden = metaData.$isTitleHidden;
			if (!this.tabTitle) {
				this.header.style.display = metaData.$isTitleHidden ? "none" : "";
			}
		}
		_applyButtonMetadata(this, metaData, designing);
		if (designing !== false) {
			if (metaData.$title !== undefined) {
				this.setTitle(this.$item.$title = metaData.$title);
			}
			designing && _applyDesignFieldsSettings(this, metaData);

			if (metaData.$field !== undefined) {
				this.$item.$field = this.$item.$field || {};
				var $props = Object.keys(metaData.$field);
				for (var ii = 0, jj = $props.length; ii < jj; ii++) {
					var $prop = $props[ii];
					this.$item.$field[$prop] = metaData.$field[$prop];
				}
			}
		}
	},
	showButton: function(id, show, syraOnClick, fontIcon) {
		_showButton(this, id, show, syraOnClick, fontIcon);
	},
	ensureButtonsVisibility: function() {
		if (this.btns && this.btns.items) {
			var display = this.$opened ? "" : "none";
			if (this.btns.items.maximize) {
				this.btns.items.maximize.style.display = display;
			}
			if (this.btns.items.refresh) {
				this.btns.items.refresh.style.display = display;
			}
		}
	},
	getArticle: function() {
		return this.boxParent ? this.boxParent.getArticle() : null;
	},
	loadBox: function() {
		this.drawBox();
	},
	getTitle: function(applyDefault) {
		var title = (this.titleLabel && this.titleLabel.textContent) || (this.$item && this.$item.$title) || "";
		if ((title == "" || title == "-") && applyDefault !== false) {
			title = this.getDefaultTitle();
		}
		return title;
	},
	isChild: function(item) {
		var boxParent = item.boxParent;
		while (boxParent && boxParent != this) {
			boxParent = boxParent.boxParent;
		}
		return boxParent == this;
	},
	setTitle: function(title, isDynamic) {
		if (this.domTitle) {
			if (isDynamic && this.$item.$XID) {
				//set only by convergence setDataContainer
				this.page.dynamicTitles = this.page.dynamicTitles || {};
				this.page.dynamicTitles[this.$item.$XID] = title;
			}
			if (title != null && title != "") {
				this.titleText = title;
				this.domTitle.style.display = "";
				if (this.titleText == "-") {
					this.titleText = "";
				}
				if (this.titleText.indexOf("{") < 0) {
					this.titleLabel.textContent = (this.$item.$noText) ? "" : this.titleText;
					if (!this.domTitle.syrainout) {
						this.domTitle.title = this.titleText;
					}
				} else {
					this.titleLabel.textContent = "";
					var parsedText = syra_site.expressionMaker.render(this, this.titleText, this.titleLabel, this.$skin + "-title-field", false);
					if (parsedText && parsedText.charAt(0) == "@") {
						this.isTitleUnlocalized = this.titleText == ("{" + parsedText + "}");
						this.domTitle.style.visibility = this.isTitleUnlocalized ? "hidden" : "";
					} else {
						this.titleText = parsedText;
						if (this.isTitleUnlocalized) {
							this.isTitleUnlocalized = false;
							this.domTitle.style.visibility = "";
						}
					}
				}
				if (this.tabTitle) {
					this.tabTitleText.textContent = this.titleText;
					this.tabTitle.title = this.titleText;
					if (this.isTitleUnlocalized !== undefined) {
						this.tabTitleText.style.visibility = this.isTitleUnlocalized ? "hidden" : "";
					}
				}
			} else {
				this.domTitle.style.display = "none";
				if (this.tabTitle) {
					this.tabTitleText.textContent = this.titleText = title = this.getDefaultTitle();
					this.tabTitle.title = title;
				}
			}
			if (!this.tabTitle) {
				this.header.style.display = (title != null && title != "") && !(this.$isTitleHidden || this.$item.$isTitleHidden) ? "" : "none";
			}
		}
	},

	_createBoxTitle: function() {
		this.titleLabel = this.domTitle = this.domTitle || document.createElement("div");
		this.domTitle.style.display = "none";
		this.domTitle.className = this.$skin + "-title";
		this.domTitle.syraItem = this.id;
		this.$item.$titleIcon && _setTitleIcon(this);
		return this.domTitle;
	},
	setDescription: function($description) {
		if ($description != null && this.header) {
			if (!this.domDescription) {
				this.domDescription = document.createElement("div");
				this.domDescription.className = this.$skin + "-description";
				this.header.appendChild(this.domDescription);
			}
			if ($description.indexOf("{") < 0) {
				this.domDescription.textContent = $description;
				if (this.tabTitle) {
					this.tabTitle.title = $description;
				}
			} else {
				syra_site.expressionMaker.render(this, $description, this.domDescription, this.$skin + "-description-field", false);
			}
		}
	},
	ensureSkin: function() {
		this.$skin = this.$skin || this.$item.$skin;
	},
	drawBox: function() {
		this.ensureSkin();
		this.domItem = document.createElement("div");

		if (this.$item.$expression) {
			if (this.$item.$css) {
				this.domItem.className = this.$item.$css;
			}
			this.layoutSlot.appendChild(this.domItem);
			this.$isEditMode = this.articleParent.$isEditMode;
			syra_site.expressionMaker.render(this, this.$item.$expression, this.domItem);
		} else {
			this.body = document.createElement("div");
			this.showBody(false);
			this.appendHeader();
			if (this.$item.$isPopupContent) {
				if (this.$item.$opened === undefined) {
					this.$item.$opened = false;
				}
			}
			if (this.isTabLayout) {
				delete this.$item.$opened;
			}
			this.ensureLayoutMode();
			if (this.$item.$width) {
				this.domItem.style.width = this.$item.$width;
			}
			if (!this.$item.$isPopupContent) {
				this.domItem.appendChild(this.body);
			}
			if (this.layoutSlot) {
				this.layoutSlot.appendChild(this.domItem);
			}
			this.setDescription(this.$item.$description);

			this.applyDesignMetaData(this.$item, false);
			this.openBox(!this.tabTitle && this.$item.$opened !== false, true, true);
		}
	},
	appendHeader: function() {
		this.header = document.createElement("header");
		this.header.className = this.$skin + "-head";
		var $title = this.$item.$title;
		if (this.$item.$XID && this.page.dynamicTitles) {
			$title = this.page.dynamicTitles[this.$item.$XID] || $title;
		}
		if (this.$item.$style) {
			if (this.$item.$style == "header") {
				if ($title == "-") {
					$title = "";
					this.$item.$isTitleHidden = true;
				}
			}
			this.header.className += " s-cst-sty-" + this.$item.$style;
		}
		this.header.style.display = "none";
		this.header.appendChild(this._createBoxTitle());
		this.domItem.appendChild(this.header);

		if (this.$field && (!$title || $title == "-")) {
			$title = this.$field.$title;
		}
		this.setTitle($title);
	},
	onOpenButtonClick: function() {
		var self = this;
		self.openBox(self.$opened);
		if (self.$item.$isPopupContent) {
			if (!self.menusPopup) {
				self.showBody(true);
				self.menusPopup = syra_site.dialogManager.openPopup(self, {
					content: self,
					slot: self.body,
					position: {
						my: "right top",
						at: "right bottom",
						of: $(self.domTitle)
					},
					onClose: function() {
						if (self.domTitle) {
							syra_site.dom.toggleClass(self.domTitle, "s-close", true);
						}
						self.$opened = false;
						self.menusPopup = null;
					}
				});
			} else {
				self.menusPopup.close();
			}
		}
	},
	expandBody: function($opened, event) {
		var self = this;
		if (event) {
			if (event.shiftKey) {
				if (self.page.isBoxesExpanded === undefined) {
					self.page.isBoxesExpanded = true;
				}
				self.page.isBoxesExpanded = !self.page.isBoxesExpanded;
				if (self.boxParent && self.boxParent.layoutContent) {
					_expandSiblingBoxes(self, self.boxParent.layoutContent, self.page.isBoxesExpanded);
				}

			}
		}
		self.$opened = ($opened === undefined) ? !self.$opened : $opened;
		self.page.externalAdapter.onBoxToggle({
			box: self,
			open: self.$opened,
			isFirstTime: !self.loaded || false,
			doEvent: function() {
				self.onOpenButtonClick();
				if (!self.$item.$isPopupContent) {
					self.articleParent.ensureArticleVisibility(true, self);
				}
			}
		});
		self.expandBtn && _setExpandButtonState(self);
		self.ensureButtonsVisibility();
	},
	renderLayoutContent: function() {
		this.layoutContent = this.page.createNewItem(this.$item.$layout, this);
		this.layoutContent.layoutSlot = this.body;
		this.layoutContent.loadBox();
	},

	openBox: function(open, load, drawing) {
		if ((open || load) && !this.loaded) {
			this.renderLayoutContent();
			this.loaded = true;
		}
		this.$opened = open;
		if (this.tabTitle) {
			if (open) {
				this.ensureLayoutMode();
			}
			if (!drawing) {
				if (open) {
					this.$item.$opened = true;
				} else {
					delete this.$item.$opened;
				}
			}
			syra_site.dom.toggleClass(this.tabTitle, "s-tab-open", open);
		}
		if (this.$item.$isBoxCollapsable) {
			if (this.expandBtn) {
				this.expandBtn && _setExpandButtonState(this);
				this.expandBtn.syraOnClick = "onOpenerClick";
			}
			this.domTitle.syraOnClick = "onOpenerClick";
			syra_site.dom.toggleClass(this.domTitle, "s-close", !open);
		}
		this.showBody(open);
		this.ensureButtonsVisibility();
	},
	showBody: function(show) {
		if (!(show && this.isBodyShow == undefined)) {
			this.isBodyShow = show;
			this.body.style.visibility = show ? "" : "hidden";
			this.body.style.height = show ? "" : "0px";
			this.body.style.width = show ? "" : "0px";
			this.body.style.borderWidth = show ? "" : "0";
			this.body.style.padding = show ? "" : "0";
			this.body.style.overflow = show ? "" : "hidden";
		}
		this.isBodyShow = show;
	},
	ensureLayoutMode: function() {
		this.ensureSkin();
		if (this.domItem) {
			var domItemCss, bodyCss, domTitleCss, headerCss;
			var $style;
			if (this.$item.$style) {
				$style = " s-cst-sty-" + this.$item.$style;
			}
			if (this.isTabLayout) {
				if (!this.tabTitle || this.isTabTitleRemoved) {
					_addTabTitle(this);
					this.tabTitle.syraItem = this.tabTitle.syrainout = this.id;
					this.isTabTitleRemoved = null;
					if (this.btns) {
						if (this.btns.slot) {
							syra_site.dom.removeChild(this.btns.slot);
						}
						this.btns = null;
						_applyButtonMetadata(this, this.$item);
					}

				}
				domItemCss = this.$skin + "-tab-box";
				bodyCss = this.$skin + "-tab-body";
				syra_site.dom.toggleClass(this.tabTitle, "s-tab-open", this.$opened);
				this.showBody(this.$opened);
				if ($style) {
					domItemCss += $style;
					bodyCss += $style;
					this.tabTitle.className += $style;
				}
			} else {
				if (this.tabTitle) {
					syra_site.dom.removeChild(this.tabTitle);
					delete this.tabTitle;
					delete this.tabTitleHead;
					delete this.tabTitleText;
					if (this.header) {
						this.header.style.display = "";
					}
					this.showBody(true);
					if (this.btns) {
						if (this.btns.slot) {
							syra_site.dom.removeChild(this.btns.slot);
						}
						this.btns = null;
						_applyButtonMetadata(this, this.$item);
					}
				}
				domItemCss = this.$skin;
				headerCss = this.$skin + "-head";
				domTitleCss = this.$skin + "-title";
				bodyCss = this.$skin + "-body";
				if (this.$item.$css) {
					var $css = " " + this.$item.$css;
					domItemCss += $css;
					bodyCss += $css;
					headerCss += $css;
					domTitleCss += $css;
				}
				if (this.$item.$isPopupContent) {
					headerCss += " s-popup";
					bodyCss += " s-popup";
					domTitleCss += " s-popup";
				}
				if (this.$item.$titleIcon) {
					if (this.$item.$titleIcon.$mode === "icon") {
						domTitleCss += " s-icon";
						headerCss += " s-icon";
					}
				}
				if ($style) {
					domItemCss += $style;
					headerCss += $style;
					domTitleCss += $style;
					bodyCss += $style;
				}
			}
			this.domItem.className = domItemCss;
			if (this.body) {
				this.body.className = bodyCss;
			}
			if (this.domTitle) {
				this.domTitle.className = domTitleCss;
			}
			if (this.header) {
				this.header.className = headerCss;
			}
			if (this.layoutParent) {
				this.layoutParent.ensureLayoutSlot(this);
			}
		}
	},
	showItem: function(show) {
		this.isItemHidden = !show;
		var display = show ? "" : "none";
		if (this.domItem) {
			this.domItem.style.display = display;
		}
		if (this.tabTitle) {
			this.tabTitle.style.display = display;
		}
		if (this.layoutParent) {
			this.layoutParent.ensureItemVisibility(this, !show);
		}
	},
	setState: function(state) {
		if (state.$isHidden !== undefined) {
			if (this.$isHidden !== state.$isHidden) {
				this.showItem(!(this.$isHidden = state.$isHidden));
			}
		}
	},
	dispose: function() {
		syra_store.remove(this);
		syra_site.dialogManager.closeAll(this);
		this.childrenSection = null;
		this.layoutContent && this.layoutContent.dispose();
		syra_site.disposeObject(this);
	}
});
