// syracuse-ui/lib/article/menuBar.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function MenuBar(){
}

exports.MenuBar = helpers.defineClass(MenuBar, null, {
    load: function(page){
        var self = this;
        self.preferences = document.site.ensurePreferences();
        self.preferences.menuBar = self.preferences.menuBar || {};
        var $menus = page.getMenuSetting();
        if ($menus) {
            page.initializeNewItem(self);
            if (self.preferences.menuBar.isDocked === undefined) {
                self.preferences.menuBar.isDocked = true;
            }
            self._slot = document.createElement("div");
            self._slot.className = self.page.$skin + "-menus-slot";
            self._slot.style.display = "none";
            self._slot.width = "150px";
            self.$$slot = $(self.page._item.appendChild(self._slot));
            
            self._bar = document.createElement("div");
            self._bar.className = "s-page-menu-bar";
            self._slot.appendChild(self._bar);
            self._appendHeader();
            
            self._body = document.createElement("div");
            self._body.className = "s-page-menu-bar-body";
            self._bar.appendChild(self._body);
            
            self._menusBox = self.page.loadNewItem($(self._body), {
                $mainAction: $menus.$mainAction,
                $bind: "$linkBox",
                $skin: self.page.$layoutOptions.menus,
                $category: "links",
                $excludeBind: $menus.$linkBox.$excludeBind,
                onAddMenuItem: function(menuItem){
                    self._slot.style.display = "";
                    // append menuIcons div
                    var div = document.createElement("div");
                    div.className = "s-page-menus-icons";
                    menuItem.$$container.append(menuItem.$$menuIcons = $(div));
                }
            });
            self._bindEvents(true);
            self._ensureState();
        }
    },
    _appendHeader: function(){
        this._header = document.createElement("header");
        this._header.className = "s-page-menu-bar-header";
        
        this.collapse = document.createElement("a");
        this.collapse.setAttribute("data-s-picker", "s-bar-collapse");
        this.collapse.className = "s-page-menu-bar-collapse";
        this._header.appendChild(this.collapse);
        
        this.dockMode = document.createElement("a");
        this.dockMode.setAttribute("data-s-picker", "s-bar-dockMode");
        this.dockMode.className = "s-page-menu-bar-docked";
        this._header.appendChild(this.dockMode);
        this._bar.appendChild(this._header);
        
        this.pickerBar = document.createElement("a");
        this.pickerBar.setAttribute("data-s-picker", "s-bar-collapse");
        this.pickerBar.className = "s-page-menu-bar-picker";
        this._bar.appendChild(this.pickerBar);
    },
    onClickPicker: function(pickerId){
        switch (pickerId) {
            case "s-bar-dockMode":
                this.preferences.menuBar.isDocked = !this.preferences.menuBar.isDocked;
                break;
            case "s-bar-collapse":
                this.preferences.menuBar.isCollapsed = !this.preferences.menuBar.isCollapsed;
                break;
        }
        this._ensureState();
    },
    _bindEvents: function($bind){
        var self = this;
        if ($bind) {
            self.$$slot.bind("mouseenter mouseleave", function(event){
                clearTimeout(self._closeTimeout);
                if (event.type == "mouseleave" && !self.preferences.menuBar.isCollapsed && !self.preferences.menuBar.isDocked) {
                    self._closeTimeout = setTimeout(function(){
                        self.preferences.menuBar.isCollapsed = !self.preferences.menuBar.isCollapsed;
                        self._ensureState();
                    }, 300);
                }
            });
        }
        else {
            if (self.$$slot) {
                self.$$slot.unbind();
            }
        }
    },
    _ensureState: function(){
        if (this.preferences.menuBar.isCollapsed) {
            this._slot.style.width = "15px";
            this._header.style.display = "none";
            this._body.style.display = "none";
            this.pickerBar.style.width = "15px";
            if (!this.preferences.menuBar.isDocked) {
                this._bar.style.width = "15px";
            }
        }
        else {
            this.pickerBar.style.width = "";
            if (this.preferences.menuBar.isDocked) {
                this.dockMode.className = "s-page-menu-bar-docked";
                this._bar.style.width = "";
                this._slot.style.width = "150px";
                this._bar.style.position = "";
            }
            else {
                this.dockMode.className = "s-page-menu-bar-float";
                this._slot.style.width = "15px";
                document.site.setZIndex(this._bar);
                this._bar.style.position = "absolute";
                this._bar.style.width = "150px";
                this._bar.style.top = "0px";
                this._bar.style.right = "0px";
            }
            this._header.style.display = "";
            this._body.style.display = "";
        }
    },
    onWindowResize: function(){
        if (this._bar && this._slot) {
            var height = document.site.body.clientHeight;
            if (height) {
                if (!this.preferences.menuBar.isCollapsed && this.preferences.menuBar.isDocked) {
                    height = document.site.body.scrollHeight;
                }
                this._bar.style.height = height + "px";
                this.pickerBar.style.height = height + "px";
                this._body.style.height = (height - this._body.offsetTop) + "px";
            }
        }
    },
    dispose: function(){
        this._bindEvents(false);
        this._menusBox = this.$$slot = this._slot = this._bar = this._header = null;
        this._body = this.collapse = this.dockMode = this.page = null;
    }
});
