// syracuse-ui/lib/article/menuBar.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function MenuBar(){
}

exports.MenuBar = helpers.defineClass(MenuBar, null, {
    load: function(page){
        this.preferences = document.site.ensurePreferences();
        this.preferences.menuBar = this.preferences.menuBar || {};
        page.initializeNewItem(this);
        if (this.preferences.menuBar.isDocked === undefined) {
            this.preferences.menuBar.isDocked = true;
        }
        this._slot = document.createElement("div");
        this._slot.className = this.page.$skin + "-menus-slot";
        this._slot.style.display = "none";
        this._slot.width = "150px";
        this.$$slot = $(this.page._item.appendChild(this._slot));
        
        this._bar = document.createElement("div");
        this._bar.className = "s-page-menu-bar";
        this._slot.appendChild(this._bar);
        this._appendHeader();
        
        this._body = document.createElement("div");
        this._body.className = "s-page-menu-bar-body";
        this._bar.appendChild(this._body);
        
        var $$body = $(this._body);
        this.page.appendMainAction($$body);
        if (this.page.$item.$menus) {
            this.page.$item.$menus.$skin = this.page.$menus.$skin || this.page.$layoutOptions.menus;
            this.page.$item.$menus.$category = "menus";
            this.page.ensureLinkSettings(this.page.$item.$menus, true);
            this.page.loadNewItem($$body, this.page.$item.$menus);
        }
        this.page.loadNewItem($$body, {
            $bind: "$linkBox",
            $skin: this.page.$layoutOptions.menus,
            $category: "links"
        });
        this._bindEvents(true);
        this._ensureState();
    },
    onMenuDataFilled: function(){
        if (!this.isSlotVisible) {
            this._slot.style.display = "";
            this.isSlotVisible = true;
        }
        
    },
    _appendHeader: function(){
        this._header = document.createElement("header");
        this._header.className = "s-page-menu-bar-header";
        
        this.collapse = document.createElement("a");
        this.collapse.setAttribute("data-s-picker", "s-bar-collapse");
        this.collapse.className = "s-page-menu-bar-collapse";
        this._header.appendChild(this.collapse);
        
        this.dockMode = document.createElement("a");
        this.dockMode.setAttribute("data-s-picker", "s-bar-dockMode");
        this.dockMode.className = "s-page-menu-bar-docked";
        this._header.appendChild(this.dockMode);
        this._bar.appendChild(this._header);
        
        this.pickerBar = document.createElement("a");
        this.pickerBar.setAttribute("data-s-picker", "s-bar-collapse");
        this.pickerBar.className = "s-page-menu-bar-picker";
        this._bar.appendChild(this.pickerBar);
    },
    onClickPicker: function(pickerId){
        switch (pickerId) {
            case "s-bar-dockMode":
                this.preferences.menuBar.isDocked = !this.preferences.menuBar.isDocked;
                break;
            case "s-bar-collapse":
                this.preferences.menuBar.isCollapsed = !this.preferences.menuBar.isCollapsed;
                break;
        }
        this._ensureState();
    },
    _bindEvents: function($bind){
        var self = this;
        if ($bind) {
            self.$$slot.bind("mouseenter mouseleave", function(event){
                clearTimeout(self._closeTimeout);
                if (event.type == "mouseleave" && !self.preferences.menuBar.isCollapsed && !self.preferences.menuBar.isDocked) {
                    self._closeTimeout = setTimeout(function(){
                        self.preferences.menuBar.isCollapsed = !self.preferences.menuBar.isCollapsed;
                        self._ensureState();
                    }, 300);
                }
            });
        }
        else {
            if (self.$$slot) {
                self.$$slot.unbind();
            }
        }
    },
    _ensureState: function(){
        if (this.preferences.menuBar.isCollapsed) {
            this._slot.style.width = "15px";
            this._header.style.display = "none";
            this._body.style.display = "none";
            this.pickerBar.style.width = "15px";
            if (!this.preferences.menuBar.isDocked) {
                this._bar.style.width = "15px";
            }
        }
        else {
            this.pickerBar.style.width = "";
            if (this.preferences.menuBar.isDocked) {
                this.dockMode.className = "s-page-menu-bar-docked";
                this._bar.style.width = "";
                this._slot.style.width = "150px";
                this._bar.style.position = "";
            }
            else {
                this.dockMode.className = "s-page-menu-bar-float";
                this._slot.style.width = "15px";
                document.site.setZIndex(this._bar);
                this._bar.style.position = "absolute";
                this._bar.style.width = "150px";
                this._bar.style.top = "0px";
                this._bar.style.right = "0px";
            }
            this._header.style.display = "";
            this._body.style.display = "";
        }
    },
    onWindowResize: function(){
        if (this._bar && this._slot) {
            var height = document.site.body.clientHeight;
            if (height) {
                if (!this.preferences.menuBar.isCollapsed && this.preferences.menuBar.isDocked) {
                    height = document.site.body.scrollHeight;
                }
                this._bar.style.height = height + "px";
                this.pickerBar.style.height = height + "px";
                this._body.style.height = (height - this._body.offsetTop) + "px";
            }
        }
    },
    dispose: function(){
        this._bindEvents(false);
        this.$$slot = this._slot = this._bar = this._header = null;
        this._body = this.collapse = this.dockMode = this.page = null;
    }
});
