// syracuse-ui/lib/article/menuBar.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function MenuBar(){
}

exports.MenuBar = helpers.defineClass(MenuBar, null, {
    ensurePreferences: function(){
        var preferences = document.site.ensurePreferences();
        return preferences.menuBar = preferences.menuBar || {};
    },
    load: function(page){
        var self = this;
        var $menus = page.getMenuSetting();
        if ($menus) {
            self.page = page;
            var preferences = self.ensurePreferences();
            if (preferences.isDocked === undefined) {
                preferences.isDocked = true;
            }
            self._slot = document.createElement("div");
            self._slot.className = self.page.$skin + "-menus-slot";
            self._slot.style.display = "none";
            self._slot.width = "150px";
            self.$$slot = $(self.page.$$item[0].appendChild(self._slot));
            
            self._bar = document.createElement("div");
            self._bar.className = "s-page-menu-bar";
            self.$$bar = $(self._slot.appendChild(self._bar));
            self._appendHeader();
            
            self._barBody = document.createElement("div");
            self._barBody.className = "s-page-menu-bar-body";
            self.$$barBody = $(self._bar.appendChild(self._barBody));
            
            self._menusBox = self.page.loadNewItem(self.$$barBody, {
                $mainAction: $menus.$mainAction,
                $bind: "$linkBox",
                $skin: self.page.$layoutOptions.menus,
                $category: "links",
                $excludeBind: $menus.$linkBox.$excludeBind,
                onAddMenuItem: function(){
                    self._slot.style.display = "";
                }
            });
            self._menusBox.$$item[0].style.width = "150px";
            
            self._bindEvents(true);
        }
    },
    _appendHeader: function(){
        var barHeader = document.createElement("header");
        barHeader.className = "s-page-menu-bar-header";
        
        this.collapse = document.createElement("a");
        this.collapse.setAttribute("data-s-picker", "collapse");
        this.collapse.className = "s-page-menu-bar-collapse";
        barHeader.appendChild(this.collapse);
        
        this.dockMode = document.createElement("a");
        this.dockMode.setAttribute("data-s-picker", "dockMode");
        this.dockMode.className = "s-page-menu-bar-docked";
        barHeader.appendChild(this.dockMode);
        
        
        this.$$barHeader = $(this._bar.appendChild(barHeader));
        
        this.expand = document.createElement("a");
        this.expand.style.display = "none";
        this.expand.setAttribute("data-s-picker", "collapse");
        this.expand.className = "s-page-menu-bar-expand";
        this._slot.appendChild(this.expand);
    },
    _bindEvents: function($bind){
        var self = this;
        if ($bind) {
            self.$$slot.delegate("a[data-s-picker]", "click", function(){
                var retValue;
                var preferences = self.ensurePreferences();
                switch (this.getAttribute("data-s-picker")) {
                    case "dockMode":
                        preferences.isDocked = !preferences.isDocked;
                        retValue = false;
                        break;
                    case "collapse":
                        preferences.isCollapsed = !preferences.isCollapsed;
                        self._isAutoCollapseDisabled = false;
                        retValue = false;
                        break;
                }
                if (retValue === false) {
                    self.onWindowResize();
                }
                return retValue;
            }).bind("mouseenter mouseleave", function(event){
                var preferences = self.ensurePreferences();
                clearTimeout(self._closeTimeout);
                if (event.type == "mouseleave") {
                    if (!self._isAutoCollapseDisabled) {
                        if (!preferences.isDocked && !preferences.isCollapsed) {
                            self._closeTimeout = setTimeout(function(){
                                preferences.isCollapsed = true;
                                self.onWindowResize();
                            }, 500);
                        }
                    }
                    delete self._isAutoCollapseDisabled;
                    
                }
                else {
                    if (!self._isAutoCollapseDisabled) {
                        if (!preferences.isDocked && preferences.isCollapsed) {
                            self._closeTimeout = setTimeout(function(){
                                preferences.isCollapsed = false;
                                self.onWindowResize();
                            }, 500);
                        }
                    }
                    delete self._isAutoCollapseDisabled;
                }
            });
        }
        else {
            if (self.$$slot) {
                self.$$slot.undelegate().unbind();
            }
        }
    },
    onWindowResize: function(){
        if (this._bar && this.$$slot) {
            var barHeight = document.site.$$body[0].clientHeight;
            if (barHeight) {
                if (this.ensurePreferences().isCollapsed) {
                    this._slot.style.width = "20px";
                    this._bar.style.display = "none";
                    this.expand.style.display = "";
                }
                else {
                    if (this.ensurePreferences().isDocked) {
                        this.dockMode.className = "s-page-menu-bar-docked";
                        this._slot.style.width = "150px";
                        this._bar.style.width = "150px";
                        this._bar.style.top = "";
                        this._bar.style.right = "";
                        this._bar.style.position = "";
                    }
                    else {
                        barHeight = document.site.$$body[0].scrollHeight;
                        this.dockMode.className = "s-page-menu-bar-float";
                        this._slot.style.width = "20px";
                        document.site.setZIndex(this._bar);
                        this._bar.style.position = "absolute";
                        this._bar.style.width = "150px";
                        this._bar.style.top = "0px";
                        this._bar.style.right = "0px";
                    }
                    this._bar.style.display = "";
                    this.expand.style.display = "none";
                }
                this._bar.style.height = barHeight + "px";
                this._barBody.style.height = (barHeight - this.$$barHeader.outerHeight(true)) + "px";
            }
        }
    },
    dispose: function(){
        //disposeOk
        this._bindEvents(false);
        delete this._menusBox;
        delete this.$$slot;
        delete this._slot;
        delete this._bar;
        delete this.$$bar;
        delete this.$$barHeader;
        delete this._barBody;
        delete this.$$barBody;
        delete this.collapse;
        delete this.dockMode;
        delete this.expand;
        delete this.page;
    }
});
