// syracuse-ui/lib/article/page.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var RawPage = require('./rawPage').RawPage;
var locale = require('syracuse-core/lib/locale');
var MenuBar = require('./menuBar').MenuBar;
var DiagnosesPanel = require('syracuse-ui/lib/site/diagnosesPanel').DiagnosesPanel;

function DesktopPage(){
}

exports.DesktopPage = helpers.defineClass(DesktopPage, RawPage, {
    getDefaultTitle: function(){
        return this._defaultTitle;
    },
    loadBox: function(initData){
        var localize = locale.resources(module)();
        switch (this.$pageCategory) {
            case "dashboard":
                this._defaultTitle = localize.p_dashboardTitle;
                this.$authoringSubType = "dashboard";
                this.$isSearchFacetEnabled = false;
                this.$autoFetch = false;
                break;
            case "portlet":
            case "vignette":
                this.$pageCategory = "vignette";
                this._defaultTitle = localize.p_vignetteTitle;
                if (this.$facet == "$demo") {
                    this.$autoFetch = false;
                }
                this.$isSearchFacetEnabled = false;
                this.$item.$isTitleHidden = true;
                break;
            default:
                this._defaultTitle = localize.p_pageTitle;
                this.$isSearchFacetEnabled = this.$isSearchFacetEnabled !== false;
                break;
        }
        RawPage.prototype.loadBox.call(this, initData);
    },
    onWindowResize: function(){
        if (this.menuBar) {
            this.menuBar.onWindowResize();
        }
        RawPage.prototype.onWindowResize.call(this);
    },
    appendArticleMenus: function(){
        if (this.$pageCategory == "vignette") {
            RawPage.prototype.appendArticleMenus.call(this);
        }
        else {
            (this.menuBar = new MenuBar()).load(this);
        }
        
    },
    drawUserProfile: function(userProfile){
        if (document.site.$item.$isUserProfileQuickViewEnabled !== false) {
            if (!this.headerUserProfile) {
                this.headerUserProfile = document.createElement("div");
                this.headerUserProfile.className = this.$skin + "-head-user-profile";
                this.$$header[0].appendChild(this.headerUserProfile);
            }
            else {
                document.site.emptyDom(this.headerUserProfile);
            }
            if (userProfile) {
                if (userProfile.selectedRole) {
                    var label = document.createElement("label");
                    label.className = this.$skin + "-head-user-role";
                    label.textContent = userProfile.selectedRole.description || "";
                    this.headerUserProfile.appendChild(label);
                }
                if (userProfile.selectedEndpoint) {
                    var label = document.createElement("label");
                    label.className = this.$skin + "-head-user-endpoint";
                    label.textContent = userProfile.selectedEndpoint.description || "";
                    this.headerUserProfile.appendChild(label);
                }
                userProfile.selectedRole && userProfile.selectedRole.description
            }
        }
    },
    appendHeader: function($$container){
        RawPage.prototype.appendHeader.call(this, $$container);
        if (this.$prototype.$description) {
            this.renderDescription(this.$prototype.$description);
        }
        var div = document.createElement("div");
        div.className = this.$skin + "-head-left";
        this.$$header.append($(div).append(this.$$header.children()));
    },
    showDiagnoses: function(message, options){
        if (!this.diagnosesPanel) {
            (this.diagnosesPanel = new DiagnosesPanel()).showDiagnoses(message, this, options);
        }
		else {
			this.diagnosesPanel.showDiagnoses(message, this, options);
		}
    },
    drawBox: function(){
        var dom;
        this.$item.$title = this.$item.$title || "{$title}";
        this.$skin = this.$item.$skin || this.$layoutOptions.$skin;
        dom = document.createElement("article");
        dom.className = this.$skin;
        this.$$item = $(dom).appendTo(this.$$container);
        
        this.dataSlot = document.createElement("div");
        this.dataSlot.className = this.$skin + "-data";
        this.$$item[0].appendChild(this.dataSlot);
        if (!this.$item.$isTitleHidden) {
            this.appendHeader($(this.dataSlot));
        }
        
        dom = document.createElement("div");
        dom.className = this.$skin + "-body";
        this.$$body = $(this.dataSlot.appendChild(dom));
        
        this._renderLayoutContent(this.$item);
        this.appendArticleMenus();
        if (this.menuBar && this.menuBar.$$bar && this.$isSearchFacetEnabled) {
            document.itemFactory.load(this.menuBar.$$bar, {
                $category: "searchFacets",
                $prototype: {},
                $bind: "$searchFacets"
            }, this);
        }
    },
    dispose: function(){
        if (this.menuBar) {
            this.menuBar.dispose();
            delete this.menuBar;
        }
		if (this.diagnosesPanel){
			document.controller.disposeObject(this.diagnosesPanel);
		}
        RawPage.prototype.dispose.call(this);
    },
});
