// syracuse-ui/lib/article/parametersHelper.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function _openDialog(article, values, $parameters, record, callback) {
	var $itemPage;
	if (!$parameters.$url) {
		var $prototype = helpers.object.clone($parameters);
		if (values) {
			Object.keys(values).forEach(function(prop) {
				$prototype[prop] = values[prop];
			});
		}
		if (article.$prototype && article.$prototype.$localization) {
			$prototype.$localization = article.$prototype.$localization;
		}
		$prototype.$title = $parameters.$title || document.site.localize.siteParametersTitle;
		$itemPage = {
			$category: "page",
			$urlParts: {
				$facet: "$edit"
			},
			$representation: {
				$prototype: $prototype
			}
		};
	}
	article.openDialog({
		$dialogMode: "modal",
		article: article,
		initData: values,
		onOpened: function(dialogPage) {
			if ($parameters && $parameters.$url) {
				Object.keys(values).forEach(function($prop) {
					var $defaultValue = values[$prop];
					if (dialogPage.boundFields && dialogPage.boundFields[$prop] && dialogPage.boundFields[$prop][0]) {
						var field = dialogPage.boundFields[$prop][0];
						field.setDataValue($defaultValue);
						field.notifyFieldChange(field.currentValue);
					}
				});
			}
		},
		onValidate: function(dialogPage) {
			if (dialogPage.validateFields()) {
				Object.keys(dialogPage.dataset).forEach(function($prop) {
					if ($prop.slice(0, 1) !== "$") {
						if (dialogPage.$prototype && dialogPage.$prototype.$properties && dialogPage.$prototype.$properties[$prop] && dialogPage.boundFields && dialogPage.boundFields[$prop] && dialogPage.boundFields[$prop][0] && dialogPage.$prototype.$properties[$prop].$type === "application/x-reference" && dialogPage.boundFields[$prop][0].currentValue && dialogPage.boundFields[$prop][0].currentValue.$uuid) {
							values[$prop] = dialogPage.boundFields[$prop][0].currentValue.$uuid;
						} else {
							values[$prop] = dialogPage.dataset[$prop];
						}
					}
				});
				callback(values);
				return true;
			}
			return false;
		},
		onClose: function(isCanceled) {
			if (isCanceled) {
				callback(null, true); //input propoerties was canceled 
			}
		},
		$itemPage: $itemPage,
		$url: $parameters.$url,
		$method: $parameters.$method || "GET"
	});
}

function _openSelectDialog(article, values, $parameters, record, callback) {
	article.openDialog({
		article: article,
		$url: ($parameters.$actions.$select || $parameters.$actions.$lookup).$url,
		onValidate: function(page) {
			if (page.selectedRecords) {
				values.$select = page.selectedRecords;
				callback(values);
				return;
			}
			callback(null, true); //input propoerties was canceled
		},
		onClose: function(isCanceled) {
			if (isCanceled) {
				callback(null, true); //input propoerties was canceled 
			}
		},
		onSelectRecord: function(selectedRecords, page) {
			page.selectedRecords = Object.keys(selectedRecords).map(function(rec) {
				return selectedRecords[rec].dataset;
			});
			return false;
		}
	});
}

exports.parseParameters = function(article, $parameters, record, callback) {
	var values = {};
	Object.keys($parameters).forEach(function(prop) {
		switch (prop) {
			case "$properties":
			case "$":
			case "$links":
			case "$actions":
				break;
			default:
				var expression = $parameters[prop];
				if (expression && typeof(expression) == "string") {
					expression = article.parseExpression(expression, record);
				}
				values[prop] = expression;
				break;
		}
	});
	if (callback) {
		if ($parameters.$actions) {
			_openSelectDialog(article, values, $parameters, record, callback);
			return;
		}
		if ($parameters.$properties) {
			_openDialog(article, values, $parameters, record, callback);
			return;
		}
		callback(values);
	}
	return values;
};

exports.formatCallBackMenuUrl = function(article, $menu, record, callback) {
	$menu.$parameters.$title = $menu.getTitle ? $menu.getTitle() : null;
	exports.parseParameters(article, $menu.$parameters, record, function(values, isCanceled) {
		if (isCanceled) {
			callback(null, isCanceled);
		} else {
			if ($menu.$sourceType && $menu.$sourceType.indexOf("{") == 0) {
				$menu.$type = article.parseExpression($menu.$sourceType, record, values);
				var $sourceMenu = article.$menus[$menu.$bind];
				if (($menu.$type.indexOf("{") != 0) && $sourceMenu && !$sourceMenu.$target) {
					$menu.$target = (($menu.$type.indexOf('json') == -1) && ($menu.$type != "html")) ? "blank" : "";
				}
			}
			callback(article.parseExpression($menu.$sourceUrl || $menu.$url, record, values));
		}
	});
};

exports.notifyActionChange = function(article, menu, record, target, request) {
	menu.$parameters.$title = menu.getTitle ? menu.getTitle() : null;
	exports.parseParameters(article, menu.$parameters, record, function(values, isCanceled) {
		if (!isCanceled) {
			var $bindAction = request[Object.keys(request)[0]];
			$bindAction.$parameters = {};
			delete values.$title;
			Object.keys(values).forEach(function(prop) {
				$bindAction.$parameters[prop] = values[prop];
			});
			article.page.notifyActionChange(menu, target, request);
		}
	});
};
