// syracuse-ui/lib/article/rawPage.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Article = require('./article').Article;
var sendBagFactory = require('./sendBagFactory');
var notifyHelper = require('./notifyHelper');
var DeltaManager = require('./deltaManager').DeltaManager;

var _$menus = {
    $search: {
        $linkBox: {
            $excludeBind: ["$create", "$details", "$edit", "$prototype", "$first", "$previous", "$next", "$last"]
        }
    },
    $cube: {
        $mainAction: "$create",
        $linkBox: {
            $excludeBind: ["$details", "$edit", "$prototype", "$first", "$previous", "$next", "$last"]
        }
    },
    $query: {
        $mainAction: "$create",
        $linkBox: {
            $excludeBind: ["$details", "$edit", "$prototype", "$first", "$previous", "$next", "$last"]
        }
    },
    $details: {
        $mainAction: "$edit",
        $linkBox: {
            $excludeBind: ["$draft", "$stylesheet"]
        }
    },
    $edit: {
        $mainAction: "$save",
        $linkBox: {
            $excludeBind: ["$draft", "$stylesheet", "$prototype"]
        }
    },
    $request: {
        $mainAction: "$invoke",
        $linkBox: {}
    },
    $dashboard: {
        $linkBox: {}
    },
    $test: {
        $linkBox: {}
    }
};

function RawPage(){
}

exports.RawPage = helpers.defineClass(RawPage, Article, {
    getMenuSetting: function(){
        return _$menus[this.$facet];
    },
    dispose: function(){
        document.site.removeResizeListener(this);
        if (this.deltaManager) {
            delete this.deltaManager.page;
            delete this.deltaManager;
        }
        delete this.articleMap;
        delete this.onAfterApplyPageDataChange;
        Article.prototype.dispose.call(this);
    },
    applyChange: function(newData){
        if (newData && newData.$isFullDelta) {
            this.$isFullDelta = true;
            //delete newData.$isFullDelta;
        }
        this.deltaManager.applyObjectDelta(this.dataset, newData);
        if (this.isPageSpyed) {
            document.site.spyGateway.spyDataset(this.dataset, newData);
        }
        Article.prototype.applyChange.call(this, newData);
        if (this.onAfterApplyPageDataChange) {
            this.onAfterApplyPageDataChange(this);
        }
        delete this.$isFullDelta;
    },
    ensureSendBag: function(article){
        return sendBagFactory.ensureSendBag(this, article);
    },
    loadBox: function(initData){
        this.page = this;
        document.site.addResizeListener(this);
        this.deltaManager = new DeltaManager();
        this.deltaManager.page = this;
        this.isDeltaMode = this.$facet == "$edit";
        this.articleMap = {};
        this.$layoutOptions = document.site.$item.$layoutOptions.page[this.$pageCategory];
        if (!this.$layoutOptions) {
            this.$layoutOptions = document.site.$item.$layoutOptions.page["page"];
        }
        this.$authoringLevel = "article";
        this.externalAdapter = this.$item.externalAdapter || document.site.externalAdapter;
        delete this.$item.externalAdapter;
        Article.prototype.loadBox.call(this, initData);
        if (this.$autoFetch = this.$autoFetch !== false) {
            this.fetch();
        }
        
    },
    appendArticleMenus: function(){
        var self = this;
        var $menus = self.getMenuSetting();
        if ($menus) {
            self._slot = document.createElement("div");
            self._slot.className = self.$skin + "-menus-slot";
            self._slot.style.display = "none";
            document.itemFactory.load($(self._slot).prependTo(self.$$item), {
                $mainAction: $menus.$mainAction,
                $bind: "$linkBox",
                $skin: self.$layoutOptions.menus,
                $category: "links",
                $excludeBind: $menus.$linkBox.$excludeBind,
                onAddMenuItem: function(){
                    self._slot.style.display = "";
                }
            }, self);
        }
    },
    onWindowResize: function(){
    
    },
    notifyReorderRecords: function(list, sourceRecord, targetRecord, $insert){
        notifyHelper.notifyReorderRecords(this, list, sourceRecord, targetRecord, $insert);
    },
    notifyActionChange: function(target, value, notifyServer){
        notifyHelper.notifyActionChange(this, target, value, notifyServer);
    },
    notifyDataChange: function(field, value){
        notifyHelper.notifyDataChange(this, field, value);
    },
    _notifyChangeToServer: function(sendBag){
        var self = this;
        if (this.isPageSpyed) {
            document.site.spyGateway.spyNotify(sendBag);
        }
        if (self.$prototype.$representationUrl) {
            sendBag.$url = self.$prototype.$representationUrl;
            sendBag.$etag = self.dataset.$etag;
            document.controller.sendRequest(self, {
                data: sendBag,
                method: "PUT",
                $etag: sendBag.$etag
            }, function(data, response, requestUrl){
                self.applyChange(data, response, requestUrl);
            });
        }
    },
    notifyChange: function($bind, value){
        var self = this;
        var data = self.ensureSendBag();
        if (self.dataset && self.dataset.$uuid) {
            data.$uuid = self.dataset.$uuid;
        }
        data[$bind] = value;
        data.$url = self.$prototype.$representationUrl;
        if (self.$prototype.$representationUrl) {
            data.$url = self.$prototype.$representationUrl;
            data.$etag = self.dataset.$etag;
            document.controller.sendRequest(self, {
                data: data,
                method: "PUT",
                $etag: data.$etag
            }, function(data, response, requestUrl){
                self.applyChange(data, response, requestUrl);
            });
        }
        else {
            delete data.$url;
            delete data.$etag;
            self.applyChange(data);
        }
    },
    
    reloadLayout: function($item){
        this.diposeContent();
        this.$item = $item;
        this.$$body.empty();
        this.loaded = false;
        this.boundFields = {};
        this.idMap = {};
        this.menuItems = {};
        this.$menus = {};
        this.menuBoxes = [];
        this.openBox(true);
    },
    
    getPage: function(){
        return this;
    },
    fetch: function(options, callback){
        if (options) {
            delete options.field;
        }
        document.controller.sendRequest(this, options, callback);
    }
});
