// syracuse-ui/lib/article/rawPage.js
"use strict";
//http://teamtreehouse.com/
var helpers = require('syracuse-core/lib/helpers');
var Article = require('./article').Article;

var _$menus = {
    $search: {
        $linkBox: {
            $excludeBind: ["$create", "$details", "$edit", "$prototype", "$first", "$previous", "$next", "$last"]
        }
    },
    $cube: {
        $mainAction: "$create",
        $linkBox: {
            $excludeBind: ["$details", "$edit", "$prototype", "$first", "$previous", "$next", "$last"]
        }
    },
    $query: {
        $mainAction: "$create",
        $linkBox: {
            $excludeBind: ["$details", "$edit", "$prototype", "$first", "$previous", "$next", "$last"]
        }
    },
    $details: {
        $mainAction: "$edit",
        $linkBox: {
            $excludeBind: ["$draft", "$stylesheet"]
        }
    },
    $edit: {
        $mainAction: "$save",
        $linkBox: {
            $excludeBind: ["$draft", "$stylesheet", "$prototype"]
        }
    },
    $dashboard: {
        $linkBox: {}
    }
};

function RawPage(){
}

exports.RawPage = helpers.defineClass(RawPage, Article, {
    loadBox: function(initData){
        this.$layoutOptions = document.site.$item.$layoutOptions.page[this.$pageCategory];
        this.$authoringLevel = "article";
        this.externalAdapter = this.$item.externalAdapter || document.site.externalAdapter;
        delete this.$item.externalAdapter;
        Article.prototype.loadBox.call(this, initData);
        if (this.$autoFetch = this.$autoFetch !== false) {
            this.fetch();
        }
    },
    _getFacetMenus: function(){
        return _$menus[this.$facet];
    },
    appendArticleMenus: function(action){
        var self = this;
        var $facetMenus = self._getFacetMenus();
        if ($facetMenus) {
            var dom = document.createElement("div");
            dom.className = self.$skin + "-menus";
            dom.style.display = "none";
            document.itemFactory.load(self.$$menus = $(dom)[action ? action : "appendTo"](self.$$item), {
                $mainAction: $facetMenus.$mainAction,
                $bind: "$linkBox",
                $skin: self.$layoutOptions.menus,
                $category: "links",
                $excludeBind: $facetMenus.$linkBox.$excludeBind,
                onAddMenuItem: function(){
                    self.$$menus[0].style.display = "";
                }
            }, self);
        }
    },
    getPage: function(){
        return this;
    },
    fetch: function(options, callback){
        document.controller.sendRequest(this, options, callback);
    }
});
