// syracuse-ui/lib/article/sendBagFactory.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function SendBag(){
}

exports.SendBag = helpers.defineClass(SendBag, null, {
    _findDataRecord: function($uuid){
        var found;
        if (this.dataRecords) {
            this.dataRecords.some(function(dataRecord){
                if (dataRecord.$uuid == $uuid) {
                    found = dataRecord;
                    return true;
                }
                return false;
            });
        }
        return found;
    },
    _ensureFullDataSet: function(list){
        var firstUpdated = this.dataRecords[0];
        this.dataRecords = list.ensureDataRecords().map(function(dataRecord){
            var saved;
            if (firstUpdated && firstUpdated.$uuid == dataRecord.$uuid) {
                delete firstUpdated.$index;
                saved = firstUpdated;
                firstUpdated = null;
            }
            else {
                saved = {
                    $uuid: dataRecord.$uuid
                };
            }
            return saved;
        });
        if (firstUpdated) {
            this.dataRecords.unshift(firstUpdated);
        }
    },
    _ensureRecord: function(record){
        var dataRecord = record.dataset;
        dataRecord = dataRecord.$singleField || dataRecord;
        var found = this._findDataRecord(dataRecord.$uuid);
        if (!found) {
            if (this.dataRecords.length > 0) {
                this._ensureFullDataSet(record.list);
                found = this._findDataRecord(dataRecord.$uuid);
                if (!found) {
                    this.dataRecords.unshift(found = {
                        $uuid: dataRecord.$uuid
                    });
                }
            }
            else {
                found = {
                    $uuid: dataRecord.$uuid
                };
                this.dataRecords.push(found);
            }
        }
        return found;
    },
    reorderRecord: function(sourceRecord, targetRecord){
        this._ensureRecord(sourceRecord);
        this._ensureRecord(targetRecord);
    },
    createRecord: function(record, index){
        var saved = {
            $uuid: record.dataset.$uuid
        };
        if (this.dataRecords.length > 0) {
            this._ensureFullDataSet(record.list);
            var found = this._findDataRecord(saved.$uuid);
            if (!found) {
                this.dataRecords.splice(index, 0, saved);
            }
        }
        else {
            saved.$index = index;
            this.dataRecords.push(saved);
        }
    },
    deleteRecord: function(record){
        this.saveRecordChange(record, "$isDeleted", true);
    },
    saveRecordChange: function(record, $bind, value){
        var found = this._ensureRecord(record);
        found[$bind] = value;
        return found;
    },
    dispose: function(){
        delete this.dataRecords;
    }
});


exports.ensureSendBag = function(page, article){
    var parent = article || page;
    if (parent.sendBag && page.dataset.$etag != parent.sendBag.$etag) {
        delete parent.sendBag;
    }
    if (!parent.sendBag) {
        parent.sendBag = new SendBag();
        parent.sendBag.$etag = page.dataset.$etag = page.dataset.$etag || 1;
        if (parent.isRecordDataset) {
            parent.sendBag.dataRecords = [];
        }
        
    }
    return parent.sendBag;
};
