// syracuse-ui/lib/authoring/palette/articlePanel.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function ArticlePanel() {}

exports.ArticlePanel = helpers.defineClass(ArticlePanel, null, {
	load: function(awContext) {
		this.awContext = awContext;
		var $links = {};
		var widths = ["100", "50,50", "33,33,33", "25,25,25,25", "20,20,20,20,20", "17,17,16,16,17,17"];
		for (var ii = 0, jj = widths.length; ii < jj; ii++) {
			$links["s-layout-type-" + (ii + 1)] = {
				$title: "" + (ii + 1),
				$widths: widths[ii]
			};
		}
		awContext.palette.applyChange({
			$links: $links
		});
	},
	dispose: function() {
		this.awContext = null;
	},
	_getRow: function(isField) {
		var $items = [];
		for (var ii = 1; ii <= 6; ii++) {
			$items.push({
				$bind: "s-layout-type-" + ii
			});
		}
		return {
			$category: "section",
			$skin: "s-aw-h2",
			$css: "s-aw-palette-layout",
			$isBoxCollapsable: true,
			$title: this.awContext.localize.aw_rowLayout,
			$isRow: true,
			$layout: {
				$items: [{
					$category: "menus",
					$title: this.awContext.localize.aw_addrow,
					$skin: "s-aw-cols",
					$isHidden: true,
					$itemSkin: "s-aw-cols-link",
					$layout: {
						$layoutType: "stack",
						$items: [{
							$css: "s-aw-cols-add-before",
							$bind: "$addRowBefore"
						}, {
							$css: "s-aw-cols-add-after",
							$bind: "$addRowAfter"
						}]
					}
				}, {
					$category: "menus",
					$title: this.awContext.localize.aw_columns,
					$skin: "s-aw-cols",
					"$itemSkin": "s-aw-cols-link",
					$layout: {
						$layoutType: "row",
						$autoSize: true,
						$items: $items
					}
				}, {
					$category: "field",
					$bind: "$rowAlign",
					$skin: "s-aw-field",
					$isEditMode: true
				}]
			}
		};
	},
	_getAppearance: function($level) {
		return {
			$category: "section",
			$skin: "s-aw-h2",
			$css: "s-aw-palette-" + $level,
			$isBoxCollapsable: true,
			$isStructure: true,
			$title: this.awContext.localize.aw_appearance,
			$layout: {
				$items: [{
					$category: "field",
					$css: "s-layout-tab-choice",
					$bind: "$boxTabChoice",
					$skin: "s-aw-field",
					$isEditMode: true,
					$format: "$button",
					$choiceLayout: "row",
					$icon: {
						$mode: "icon",
						$path: "authoring/s-layout-"
					}
				}]
			}
		};
	},
	_getLayout: function($level) {
		var $items = [];
		for (var ii = 1; ii <= 6; ii++) {
			$items.push({
				$bind: "s-layout-type-" + ii,
				$css: $level == "field" ? "" : "s-aw-row-box"
			});
		}
		var $layout = {
			$category: "section",
			$skin: "s-aw-h2",
			$css: "s-aw-palette-" + $level,
			$isBoxCollapsable: true,
			$isStructure: true,
			$title: this.awContext.localize.aw_layout,
			$layout: {
				$items: [{
					$category: "menus",
					$title: this.awContext.localize.aw_columns,
					$skin: "s-aw-cols",
					$itemSkin: "s-aw-cols-link",
					$layout: {
						$layoutType: "row",
						$autoSize: true,
						$items: $items
					}
				}, {
					$category: "field",
					$isHidden: !($level == "article" || $level == "section"),
					$css: "s-layout-tab-choice",
					$bind: "$layoutTabChoice",
					$skin: "s-aw-field",
					$isEditMode: true,
					$format: "$button",
					$choiceLayout: "row",
					$icon: {
						$mode: "icon",
						$path: "authoring/s-layout-"
					}
				}]
			}
		};
		return $layout;
	},
	_getModels: function($level) {
		this.awContext.palette.applyChange({
			$links: {
				$defaultLayout: {
					$title: this.awContext.localize.aw_defaultLayout
				},
				"pattern-headerTabs": {
					$title: this.awContext.localize.aw_autoLayout
				},
				"pattern-1": {
					$title: this.awContext.localize.aw_layoutComposite
				},
				"pattern-2": {
					$title: this.awContext.localize.aw_layoutComposite
				},
				"pattern-3": {
					$title: this.awContext.localize.aw_layoutComposite
				},
				"pattern-4": {
					$title: this.awContext.localize.aw_layoutComposite
				}
			}
		});
		return {
			$category: "section",
			$skin: "s-aw-h2",
			$isBoxCollapsable: true,
			$css: "s-aw-palette-" + $level,
			$title: this.awContext.localize.aw_models,
			$layout: {
				$items: [{
					$category: "menus",
					$itemIcon: {
						$mode: "icon",
						$path: "authoring/s-aw-"
					},
					$skin: "s-aw-menus",
					$itemSkin: "s-aw-menus-link",
					$layout: {
						$items: [{
							$layoutType: "row",
							$items: [{
								$icon: null,
								$bind: "pattern-headerTabs",
								$css: "s-aw-pattern"
							}, {
								$icon: null,
								$bind: "$defaultLayout",
								$css: "s-aw-pattern"
							}]

						}, {
							$layoutType: "row",
							$items: [{
								$bind: "pattern-1",
								$css: "s-aw-pattern"
							}, {
								$bind: "pattern-2",
								$css: "s-aw-pattern"
							}, {
								$bind: "pattern-3",
								$css: "s-aw-pattern"
							}, {
								$bind: "pattern-4",
								$css: "s-aw-pattern"
							}]
						}]
					}
				}]
			}
		};
	},
	get_article: function() {
		return {
			$category: "section",
			$skin: "s-aw-h1",
			$css: "s-aw-palette-article",
			$title: this.awContext.localize.aw_page,
			$isHidden: true,
			$layout: {
				$items: [{
					$category: "section",
					$isTitleHidden: true,
					$skin: "s-aw-palette-item",
					$layout: {
						$layoutType: "stack",
						$items: [this._getRow(), this._getLayout("article"), {
							$category: "section",
							$skin: "s-aw-h2",
							$css: "s-aw-palette-article",
							$bind: "$field",
							$isSectionFields: true,
							$isBoxCollapsable: true,
							$title: this.awContext.localize.aw_sectionFields,
							$layout: {
								$layoutType: "stack",
								$items: [{
									$bind: "$fieldsIsTitleHidden",
									$skin: "s-aw-field",
									$isEditMode: true,
									$isLeftSpaceHidden: true
								}, {
									$bind: "$labelAlignment",
									$skin: "s-aw-field",
									$isEditMode: true,
									$isLeftSpaceHidden: true,
									$choiceLayout: "1"
								}]
							}
						},
						this._getModels("article")]
					}
				}]
			}
		};
	},
	get_section: function() {
		return {
			$category: "section",
			$skin: "s-aw-h1",
			$css: "s-aw-palette-section",
			$title: this.awContext.localize.aw_section,
			$isHidden: true,
			$layout: {
				$items: [{
					$category: "section",
					$id: "s-aw-palette-section",
					$isTitleHidden: true,
					$skin: "s-aw-palette-item",
					$layout: {
						$layoutType: "stack",
						$items: [{
							$bind: "$excludeMe",
							$category: "link",
							$skin: "s-aw-menus-link",
							$css: "s-aw-exclude-me"
						},
						this._getRow(), this._getAppearance("section"), this._getLayout("section"), {
							$category: "section",
							$skin: "s-aw-h2",
							$css: "s-aw-palette-section",
							$isBoxCollapsable: true,
							$title: this.awContext.localize.aw_sectionAppearance,
							$layout: {
								$items: [{
									$bind: "$title",
									$skin: "s-aw-field",
									$isEditMode: true,
									$isLeftSpaceHidden: true
								}, {
									$bind: "$isTitleHidden",
									$css: "s-aw-field-is-title-hidden",
									$skin: "s-aw-field",
									$isEditMode: true
								}]
							}
						}, {
							$category: "section",
							$skin: "s-aw-h2",
							$css: "s-aw-palette-section",
							$isBoxCollapsable: true,
							$title: this.awContext.localize.aw_sectionBehaviour,
							$layout: {
								$items: [{
									$bind: "$isBoxCollapsable",
									$skin: "s-aw-field",
									$isEditMode: true,
									$isLeftSpaceHidden: true
								}, {
									$bind: "$isMaximizable",
									$skin: "s-aw-field",
									$isEditMode: true,
									$isLeftSpaceHidden: true
								}]
							}
						}, {
							$category: "section",
							$skin: "s-aw-h2",
							$css: "s-aw-palette-section",
							$isBoxCollapsable: true,
							$bind: "$field",
							$isSectionFields: true,
							$title: this.awContext.localize.aw_sectionFields,
							$layout: {
								$layoutType: "stack",
								$items: [{
									$bind: "$fieldsIsTitleHidden",
									$skin: "s-aw-field",
									$isEditMode: true,
									$isLeftSpaceHidden: true
								}, {
									$bind: "$labelAlignment",
									$skin: "s-aw-field",
									$isEditMode: true,
									$isLeftSpaceHidden: true,
									$choiceLayout: "1"
								}]
							}
						}]
					}
				}]
			}
		};
	},
	get_block: function(awItem) {
		return {
			$category: "section",
			$skin: "s-aw-h1",
			$css: "s-aw-palette-block",
			$title: this.awContext.localize.aw_block,
			$isHidden: true,
			$layout: {
				$items: [{
					$category: "section",
					$id: "s-aw-palette-block",
					$isTitleHidden: true,
					$skin: "s-aw-palette-item",
					$layout: {
						$layoutType: "stack",
						$items: [{
							$bind: "$excludeMe",
							$category: "link",
							$skin: "s-aw-menus-link",
							$css: "s-aw-exclude-me"
						},
						this._getRow(), this._getAppearance("block"), this._getLayout("block"), {
							$category: "section",
							$skin: "s-aw-h2",
							$css: "s-aw-palette-block",
							$isBoxCollapsable: true,
							$title: this.awContext.localize.aw_sectionAppearance,
							$layout: {
								$items: [{
									$bind: "$title",
									$skin: "s-aw-field",
									$isEditMode: true,
									$isLeftSpaceHidden: true
								}, {
									$bind: "$isTitleHidden",
									$css: "s-aw-field-is-title-hidden",
									$skin: "s-aw-field",
									$isEditMode: true
								}]
							}
						}, {
							$category: "section",
							$skin: "s-aw-h2",
							$css: "s-aw-palette-block",
							$isBoxCollapsable: true,
							$title: this.awContext.localize.aw_sectionBehaviour,
							$layout: {
								$items: [{
									$bind: "$isBoxCollapsable",
									$skin: "s-aw-field",
									$isEditMode: true,
									$isLeftSpaceHidden: true
								}, {
									$bind: "$isMaximizable",
									$skin: "s-aw-field",
									$isEditMode: true,
									$isLeftSpaceHidden: true
								}]
							}
						}, {
							$category: "section",
							$skin: "s-aw-h2",
							$css: "s-aw-palette-block",
							$isBoxCollapsable: true,
							$bind: "$field",
							$isSectionFields: true,
							$title: this.awContext.localize.aw_sectionFields,
							$layout: {
								$layoutType: "stack",
								$items: [{
									$bind: "$fieldsIsTitleHidden",
									$skin: "s-aw-field",
									$isEditMode: true,
									$isLeftSpaceHidden: true
								}, {
									$bind: "$labelAlignment",
									$skin: "s-aw-field",
									$isEditMode: true,
									$isLeftSpaceHidden: true,
									$choiceLayout: "1"
								}]
							}
						}]
					}
				}]
			}
		};
	},
	get_field: function() {
		return {
			$category: "section",
			$skin: "s-aw-h1",
			$css: "s-aw-palette-field",
			$title: this.awContext.localize.aw_field,
			$isHidden: true,
			$layout: {
				$items: [{
					$category: "section",
					$id: "s-aw-palette-field",
					$isTitleHidden: true,
					$skin: "s-aw-palette-item",
					$layout: {
						$layoutType: "stack",
						$items: [{
							$bind: "$excludeMe",
							$category: "link",
							$skin: "s-aw-menus-link",
							$css: "s-aw-exclude-me"
						},
						this._getRow(), this._getLayout("field"), {
							$category: "section",
							$css: "s-aw-palette-field",
							$skin: "s-aw-h2",
							"$isLabelSection": true,
							$isBoxCollapsable: true,
							$title: this.awContext.localize.aw_sectionFieldLabel,
							$layout: {
								$layoutType: "stack",
								$items: [{
									$bind: "$isTitleHidden",
									$title: this.awContext.localize.aw_isLabelHidden,
									$skin: "s-aw-field",
									$isEditMode: true,
									$isLeftSpaceHidden: true
								}, {
									$bind: "$labelAlignment",
									$title: this.awContext.localize.aw_sectionFieldLabelPosition,
									$skin: "s-aw-field",
									$isEditMode: true,
									$choiceLayout: "1"
								}]
							}
						}, {
							$category: "section",
							$title: this.awContext.localize.aw_sectionFieldWidget,
							$isBoxCollapsable: true,
							$css: "s-aw-palette-field",
							$skin: "s-aw-h2",
							$isHidden: true,
							$isFieldAddinSection: true,
							$layout: {
								$items: []
							}
						}]
					}
				}]
			}
		};
	},
	get_menu: function() {
		return {
			$category: "section",
			$skin: "s-aw-h1",
			$css: "s-aw-palette-field",
			$title: this.awContext.localize.aw_menuItem,
			$isHidden: true,
			$layout: {
				$items: [{
					$category: "section",
					$isTitleHidden: true,
					$skin: "s-aw-palette-item",
					$layout: {
						$layoutType: "stack",
						$items: [{
							$bind: "$excludeMe",
							$category: "link",
							$skin: "s-aw-menus-link",
							$css: "s-aw-exclude-me"
						},
						this._getRow(), this._getLayout("field")]
					}
				}]
			}
		};
	}
});
