// syracuse-ui/lib/site/tools/dragDropManager.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function DragDropManager() {}

exports.DragDropManager = helpers.defineClass(DragDropManager, null, {
	load: function() {
		this._observers = {
			page: [],
			colResizer: [],
			barResizer: []
		};
	},
	setDragSpot: function(dom, enable) {
		if (enable) {
			dom.setAttribute("data-s-drag-spot", "1");
		} else {
			dom.removeAttribute("data-s-drag-spot");
		}
	},
	togglePageObserver: function(observer, add) {
		this._toggleObserver("page", observer, add);
	},
	toggleColResizerObserver: function(observer, add) {
		this._toggleObserver("colResizer", observer, add);
	},
	toggleBarResizerObserver: function(observer, add) {
		this._toggleObserver("barResizer", observer, add);
	},
	_toggleObserver: function(key, observer, add) {
		if (add) {
			this._observers[key].push(observer);
		} else {
			var ii = this._observers[key].indexOf(observer);
			if (ii >= 0) {
				this._observers[key].splice(ii, 1);
			}
		}
	},
	cancel: function() {
		delete this.dropableItem;
	},
	start: function(observer, DragDropClass, boundary) {
		if (this.dropableItem) {
			this.ddAgent = new DragDropClass();
			this.ddAgent.start(this.dropableItem, observer, boundary);
			delete this.dropableItem;
		}
	},
	stop: function() {
		this.cancel();
		if (this.ddAgent) {
			if (this.ddAgent.stop) {
				this.ddAgent.stop();
			}
			this.ddAgent.dispose();
			delete this.ddAgent;
		}
		document.site.body.style.cursor = "default";
	},
	dispose: function() {
		delete this._observers;
	}
});
