// syracuse-ui/lib/site/tools/itemHelper.js
"use strict";

var _pagesMap = {};

exports.getLoadedPages = function() {
	var pages = [];
	var ids = Object.keys(_pagesMap);
	for (var ii = 0, jj = ids.length; ii < jj; ii++) {
		var page = _pagesMap[ids[ii]];
		if (page && !page.disposed) {

			pages.push(page);
		}
	}
	return pages;
};

//for convergence
exports.getDashBoard = function() {
	var categories = ["dashboard", "navigation", "landing"];
	var keys = Object.keys(_pagesMap);
	for (var ii = 0, jj = keys.length; ii < jj; ii++) {
		var page = _pagesMap[keys[ii]];
		if (categories.indexOf(page.$pageCategory) >= 0) {
			return page;
		}
	}
};

exports.findArticle = function(dom) {
	while (dom != null) {
		if (dom.syraarticle) {
			return repository[dom.syraarticle];
		}
		dom = dom.parentNode;
	}
};
exports.findField = function(dom) {
	var field;
	while (dom != null) {
		if (dom.syraItem) {
			field = repository[dom.syraItem];
			if (!field || (field && field.isField)) {
				break;
			}
		}
		dom = dom.parentNode;
	}
	return field;
};

exports.registerArticle = function(article) {
	if (article.page == article) {
		_pagesMap[article.id] = article;
	} else {
		if (article && article.page) {
			article.page._articles.push(article);
		}
	}
};
exports.unregisterArticle = function(article) {
	if (article.page) {
		if (article.page == article) {
			delete _pagesMap[article.id];
		} else {
			if (article.page._articles) {
				article.page._articles.splice(article.page._articles.indexOf(article), 1);
			}
		}
	}
};
exports.dispose = function() {
	_pagesMap = null;
};
