// syracuse-ui/lib/site/tools/itemHelper.js
"use strict";

var _pagesMap = {};

//for convergence
exports.getDashBoard = function(){
    var categories = ["dashboard", "navigation", "landing"];
    var keys = Object.keys(_pagesMap);
    for (var ii = 0, jj = keys.length; ii < jj; ii++) {
        var page = _pagesMap[keys[ii]];
        if (categories.indexOf(page.$pageCategory) >= 0) {
            return page;
        }
    }
};

exports.findPage = function(pageId){
    return _pagesMap[pageId];
}

exports.findArticle = function(dom){
    while (dom != null && !dom.syraArticle) {
        dom = dom.parentNode;
    };
    if (dom && dom.syraArticle) {
        var page = _pagesMap[dom.syraPageId];
        if (page) {
            return dom.syraPageId == dom.syraArticle ? page : page.findChildArticle(dom.syraArticle);
        }
    }
    return null;
};
exports.findField = function(dom){
    var article = exports.findArticle(dom);
    if (article && article.idMap) {
        while (dom != null && !dom.syra_field_id && !dom.syraArticle) {
            dom = dom.parentNode;
        }
        if (dom && dom.syra_field_id) {
            return article.idMap[dom.syra_field_id];
        }
    }
    return (article && (!article.idMap || !article.idMap.length) && article.isField) ? article : null;
};

exports.findBox = function(dom){
    var article = exports.findArticle(dom);
    var id = dom.syraBoxId || dom.getAttribute("data-s-box");
    while (dom != null && !id) {
        id = dom.getAttribute("data-s-box");
        dom = dom.parentNode;
    }
    var box = null;
    if (article) {
        box = article.idMap[id];
    }
    if (!box && id == article.id) {
        box = article;
    }
    return box;
};

exports.findMenu = function(dom){
    var article = exports.findArticle(dom);
    if (article) {
        var id = dom.getAttribute("data-s-menu");
        return (article.id == id) ? article : (article.idMap[id] || null);
    }
    return null;
};

exports.registerArticle = function(article){
    if (article.page == article) {
        _pagesMap[article.id] = article;
    }
    else {
        if (article && article.page) {
            article.page._articles.push(article);
        }
    }
};
exports.unregisterArticle = function(article){
    if (article.page) {
        if (article.page == article) {
            delete _pagesMap[article.id];
        }
        else {
            if (article.page._articles) {
                article.page._articles.splice(article.page._articles.indexOf(article), 1);
            }
        }
    }
};
exports.dispose = function(){
    _pagesMap = null;
};
