// syracuse-ui/lib/site/tools/maximizer.js
"use strict";

function _toggleTabSiblings(section, show) {
	var node = section.header.parentNode.firstChild;
	while (node) {
		if (node != section.header) {
			syra_dom.hide(node, !show);
		}
		node = node.nextSibling;
	}
}

function _getTopParentPage(page) {
	if (page.vignetteField) {
		page = _getTopParentPage(page.vignetteField.page);
	}
	if (page.inlinePageHost) {
		page = _getTopParentPage(page.inlinePageHost.record.page);
	}
	return page;
}

function _applyToBar(bar, maximized) {
	if (bar) {
		if (maximized) {
			bar.onBeforeMaximized = bar.$isCollapsed;
			bar.$isCollapsed = true;
			bar.ensureState();
		} else {
			if (bar.onBeforeMaximized !== undefined) {
				bar.$isCollapsed = bar.onBeforeMaximized;
				bar.ensureState();
			}
		}
	}
}

function _maximizeItem(btn) {
	var item = btn.parent;
	var page = _getTopParentPage(item.page);
	var slot = page.scrollview;
	item.isMaximized = !item.isMaximized;
	if (item.isMaximized) {
		item._parentChildren = [];
		var nodes = slot.children;
		item.siblingBeforeMaximized = item.domItem.nextSibling;
		while (nodes.length) {
			var node = nodes[0];
			node.parentNode && node.parentNode.removeChild(node);
			item._parentChildren.push(node);
		}
		slot.appendChild(item.domItem);
	} else {
		item.layoutSlot.appendChild(item.domItem);
		syra_dom.empty(slot);
		for (var ii = 0, jj = item._parentChildren.length; ii < jj; ii++) {
			slot.appendChild(item._parentChildren[ii]);
		}
		if (item.siblingBeforeMaximized) {
			item.layoutSlot.insertBefore(item.domItem, item.siblingBeforeMaximized);
			delete item.siblingBeforeMaximized;
		}
		delete item._parentChildren;
	}
	_applyToBar(page.menuBar, item.isMaximized);
	if (page.fusionBar) {
		item.$item && item.$item.$isNavigationList && page.fusionBar.toggleBar(!item.isMaximized);
		_applyToBar(page.fusionBar, item.isMaximized);
	}
	if (syra_site.landingPage) {
		syra_site.landingPage.resizeArticle(true);
	} else {
		item.page.resizeArticle(true);
	}
	var state = item.isMaximized ? "minimize" : "maximize";
	syra_menus.button.setText(btn, syra_local["box_" + state], state);
}

function _onClick() {
	_maximizeItem(this);
	if (this.parent.isSection) {
		var section = this.parent;
		if (section.isTabSection) {
			if (section.isMaximized) {
				var slot = section.domItem.parentNode;
				slot.appendChild(section.layoutParent._tabs);
				_toggleTabSiblings(section, false);
				section.openTab();

			} else {
				if (section.layoutParent.domItem.firstChild) {
					section.layoutParent.domItem.firstChild.parentNode.insertBefore(section.layoutParent._tabs, section.layoutParent.domItem.firstChild);
				} else {
					section.layoutParent.domItem.appendChild(section.layoutParent._tabs);
				}
				_toggleTabSiblings(section, true);

			}
		}
	}
	this.parent.onAfterMaximize && this.parent.onAfterMaximize();
}

exports.toggleButton = function(parent, show, slot, css) {
	if (show) {
		if (!parent.maximizeBtn) {
			parent.maximizeBtn = syra_menus.button.add({
				parent: parent,
				slot: slot,
				text: syra_local.box_maximize,
				css: css || "s-aw-list-mode",
				iconOnly: true,
				fontIcon: "maximize",
				btnclick: _onClick
			});
		}
	} else {
		syra_menus.button.remove(parent.maximizeBtn);
		delete parent.maximizeBtn;
	}
};
