// syracuse-ui/lib/site/tools/officeHelper.js
"use strict";
var _helpers = require('syracuse-core/lib/helpers');

exports.officeAddinSetupInstallation = "/msoffice/lib/general/addIn/SyracuseOfficeAddinsSetup.EXE";

exports.officeAddinInstalled = function(controller, object, article, title, callback, args) {
	if (object.$confirm && object.$officeAddinSetup) {
		if (_helpers.getCookie("SyracuseOfficeAddin2") === "valid") {
			return true;
		}
	}

	if (object.$url && object.$url.indexOf("SyracuseOfficeAddinsSetup.EXE") > 0) {
		_helpers.setCookie("SyracuseOfficeAddin2", "valid", null, (new Date((new Date().getYear() + 1910), new Date().getMonth(), new Date().getDate())).toUTCString());
	}

	if (object.$confirm) {
		setTimeout(function() {
			syra_site.closePopups();
			syra_diagnose.box.show({
				$isAutoClose: 10000,
				$title: syra_expression.parse(article, title || syra_local.siteDefaultConfirmTitle),
				$message: syra_expression.parse(article, syra_local.installOfficeAddin), // we don't use object.$confirm here to avoid using two sources of the message
				$type: "question",

				callback: function(response, closureMode) {
					if (closureMode == "auto" || (response && response.$clientId == "yes")) {
						if (response && response.$clientId == "yes" && object.$officeAddinSetup) {
							_helpers.setCookie("SyracuseOfficeAddin2", "valid", null, (new Date((new Date().getYear() + 1910), new Date().getMonth(), new Date().getDate())).toUTCString());
						}
						callback.apply(controller, args);
					}
				}
			});
		}, 50);
		return false;
	}
	return true;
};
