// syracuse-ui/lib/site/tools/pageLoader.js
"use strict";
var UserPreferences = require('syracuse-ui/lib/page/utility/userPreferences').UserPreferences;

var _childPages = [];

var _queryFacets = ["$query", "$bulk", "$lookup", "$select", "$cube"];
var _autoFetchFacets = _queryFacets.concat(["$details", "$edit", "$search", "$summary"]);


exports.getLoaded = function() {
	return _childPages;
};

exports.initialize = function(page) {
	if (!page.isSiteRegisterDisabled) {
		if (_childPages.indexOf(page) < 0) {
			_childPages.push(page);
		}
	}
	page.page = page;

	page._widgetCategories = syra_site.siteOptions.widgetsLibrary.categories;
	page._childItemOffset = page._childItemOffset || 1; //initialize for site
	if (!page.id) {
		page.id = syra_site._childPageOffset++;
		syra_store.add(page);
	}
};

exports.resize = function() {
	for (var ii = 0, jj = _childPages.length; ii < jj; ii++) {
		_childPages[ii].resizeArticle(true);
	}
};



exports.load = function($itemPage, onMainPagechange) {
	var $representation = $itemPage.$representation;
	var $pageCategory = ($representation.$article ? $representation.$article.$category : null) || $itemPage.$category;
	$itemPage.openerUrlSegments = $itemPage.openerUrlSegments || {};
	if (!$pageCategory) {
		switch ($itemPage.openerUrlSegments.representationRoot) {
			case "landingVignetteSelect":
				$pageCategory = "landingVignetteSelect";
				$itemPage.$autoFetch = true;
				break;
			default:
				switch ($itemPage.openerUrlSegments.$facet) {
					case "$dashboard":
						$pageCategory = "dashboard";
						break;
					case "$navigation_edit":
					case "$navigation":
						$pageCategory = "navigation";
						$itemPage.initData = $representation;
						break;
					case "$landing_edit":
					case "$landing":
						$pageCategory = "landing";
						$itemPage.initData = $representation;
						break;
					default:
						$pageCategory = "page";
						break;
				}
				break;
		}
	}
	var item = new($itemPage.$pageCategoryClass || syra_site.siteOptions.widgetsLibrary.pageCategories[$pageCategory] || syra_site.siteOptions.widgetsLibrary.defaultPageCategory)();
	var $userPreferencesLink = $representation.$links && $representation.$links.$userPreferences;
	if ($userPreferencesLink) {
		(item.userPreferences = new UserPreferences()).load(item, $userPreferencesLink, $representation.userPreferences && $representation.userPreferences.content);
	}
	item.id = syra_site._childPageOffset++;
	item.openerUrlSegments = $itemPage.openerUrlSegments;
	item.$pageCategory = $pageCategory;
	item.dialogWrapper = $itemPage.dialogWrapper;
	item.$pageCollaborationUrl = $representation.$pageCollaborationUrl;
	item.$authorUrl = $representation.$authorUrl;
	item.$facet = $itemPage.$facet || $itemPage.openerUrlSegments.$facet;
	item.$isEditMode = $itemPage.$isEditMode || item.$facet == "$edit";
	if ($itemPage.vignetteField) {
		item.vignetteField = $itemPage.vignetteField;
		delete $itemPage.vignetteField;
	}
	var hackFusion = item.$facet.split("_");
	if (hackFusion.length > 1 && hackFusion[0] == "$fusion") {
		item.$isEditMode = true;
		item.$facet = hackFusion[1];
		item.$fusionPageMeta = $itemPage.$fusionPageMeta;
		$itemPage.$autoFetch = false;
	}
	item.layoutSlot = $itemPage.layoutSlot;
	item.$prototype = $representation.$prototype;
	item.page = item;
	var $protoArticle = item.$prototype.$article;
	var $article = (item.initializeArticle) ? item.initializeArticle($itemPage) : $representation.$article;
	if ($article) {
		delete $article.$menus;
	} else {
		$article = $protoArticle;
	}
	$article = item.ensureDefaultArticle($article, item.$prototype);
	if (item.$prototype.$menus) {
		$article.$menus = item.$prototype.$menus;
	} else {
		if ($protoArticle && $protoArticle.$menus) {
			$article.$menus = $protoArticle.$menus;
		}
	}
	/*if (item.$prototype && item.$prototype.$localization && item.$prototype.$hackLocalization) {
     localizationConverter.convertLocalization(item.$prototype, $article);
     }*/
	var layoutSlotStyle = $itemPage.layoutSlot && $itemPage.layoutSlot.style && $itemPage.layoutSlot.style.display;

	syra_site.initializeNewItem(item, $article, $itemPage.boxParent);
	syra_site.localizer.applyPageLocalization(item);

	item.$autoFetch = $itemPage.$autoFetch;
	if (item.$autoFetch === undefined) {
		if (item.$facet == "$edit" && $itemPage.initData) {
			item.$autoFetch = false;
		} else {
			item.$autoFetch = _autoFetchFacets.indexOf(item.$facet) >= 0;
		}
	}

	item.$item.externalAdapter = $itemPage.externalAdapter;
	if ($itemPage.openerUrlSegments) {
		item.$urlParams = $itemPage.openerUrlSegments.params;
	}
	item.$views = $representation.$views;
	item.loadBox($itemPage.initData);
	if ($itemPage.onAfterLoadPage) {
		$itemPage.onAfterLoadPage(item);
	}
	item.isPageLoaded = true;
	if ($representation.$diagnoses && $representation != $itemPage.initData) {
		item.showDiagnoses({
			$diagnoses: $representation.$diagnoses
		});
	} else {
		item.diagnosesPanel && item.diagnosesPanel.refresh();
	}
	$itemPage.$pageCategoryClass = $itemPage.boxParent = $itemPage.layoutSlot = $representation = $itemPage.onAfterLoadPage = $itemPage.dialogWrapper = null;
	if (layoutSlotStyle !== undefined) {
		item.layoutSlot.style.display = layoutSlotStyle;
	}
	if (onMainPagechange) {
		syra_site.siteFunctions.refreshMainPageSecurity(item);
		syra_site.body.style.display = "";
	}
	item.ensurePageVisibility(true);
	return item;
};

exports.activateQueryList = function(page) {
	if (!page.isFusionPage && !page.isVignettePage && _queryFacets.indexOf(page.$facet) >= 0) {
		var list = syra_store.findList(page);
		if (list) {
			list.isPageActivationHidden = true;
			page.setActivateList(list);
			if (list.filterRecord) {
				for (var ii = 0, jj = list.orderCols.length; ii < jj; ii++) {
					var col = list.orderCols[ii];
					if (col.$bind) {
						var $field = list.$fields[col.$bind];
						if ($field && !$field.$isHidden && !$field.$isHidden && !$field.$isDisabled) {
							var cell = list.filterRecord._cells[col.$bind];
							if (cell && cell.field) {
								cell.field.focus();
								break;
							}
						}
					}
				}
			}
		}
	}
};

exports.disposePage = function(page) {
	if (page.userPreferences) {
		page.userPreferences.dispose();
	}
	if (!page.isVignettePage) {
		var found = _childPages.indexOf(page);
		if (found >= 0) {
			_childPages.splice(found, 1);
		}
	}
	page.userPreferences = null;
};
