// syracuse-ui/lib/site/tools/siteFunctions.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var bookmarksManager = require("syracuse-ui/lib/page/bookmarks/bookmarksManager");
var _dragDropManager = require("syracuse-ui/lib/site/tools/dragDropManager");


exports.addBookmarkManager = function() {
	syra_site.bookmarksManager = bookmarksManager;
	syra_site.bookmarksManager.load();
};

exports.addJobsOpener = function(slot) {
	var btn = syra_site._jobsViewerOpener = syra_menus.addIconButton("", "s-jobs-viewer-opener", "onJobsOpenerClick");
	btn.syraSiteObserver = "siteFunctions";
	btn.style.display = "none";
	slot.appendChild(btn);
};
exports.onJobsOpenerClick = function() {
	syra_site.jobsPage && syra_site.jobsPage.toggle();
};

function _runJobsViewer() {
	syra_controller.callServer(null, {
		method: "GET",
		$location: {
			$url: "/sdata/$trackers"
		}
	}, function(data, response, $url) {
		if (data.$resources && data.$resources.length > 0) {
			for (var ii = 0, jj = data.$resources.length; ii < jj; ii++) {
				var resource = data.$resources[ii];
				var params = {
					$location: resource.location || resource.$location,
					$state: resource.phase,
					$title: resource.title || "(Title not retrieved)",
					uuid: resource.uuid,
					kind: "operation"
				};
				syra_site.registerJobSyra(params.$location, params.$state, params.$title, params.uuid);
			}
		}
	}, function(error, httpquery) {
		httpquery.showError(error);
	});
}

exports.addLogo = function(slot) {
	var title, href;
	var home = syra_site.menuItems && syra_site.menuItems.$home;
	home = home && home[0];
	if (home) {
		title = home.domItem.title;
		href = home.domItem.getAttribute("href");
	}
	var btn = syra_menus.addIconButton(title || "", "s-site-header-logo", "onLogoClick");
	btn.syraSiteObserver = "siteFunctions";
	href && btn.setAttribute("href", href);
	slot.appendChild(btn);

};
exports.onLogoClick = function() {
	syra_menus.click.home();
};

exports.addLogoutButton = function() {
	var btn = syra_menus.addIconButton(syra_local.siteLogoutLink, "s-site-logout", "onLogoutClick");
	btn.syraSiteObserver = "siteFunctions";
	syra_site.headerTop.appendChild(btn);
};
exports.onLogoutClick = function() {
	syra_site.logout();
};

exports.addDragDropManager = function() {
	window.syra_dd = _dragDropManager;
};

exports.refreshMainPageSecurity = function(page) {
	if (page && !page.disposed) {
		var role = syra_site.userProfile.getSelectedRole();
		var endpoint = syra_site.userProfile.getSelectedEndpoint();
		if (role || endpoint) {
			role = role && role.description;
			endpoint = endpoint && endpoint.description;
			var pages = page ? [page] : syra_site.pageLoader.getLoaded();
			for (var ii = 0, jj = pages.length; ii < jj; ii++) {
				page = pages[ii];
				if (page.securityViewSlot) {
					syra_site.dom.empty(page.securityViewSlot);
					role && page.securityViewSlot.appendChild(syra_menus.addFontIconText(role, "s-security-user", "role"));
					endpoint && page.securityViewSlot.appendChild(syra_menus.addFontIconText(endpoint, "s-security-user", "endpoint"));
				}
			}
		}
		if (page.mainPageDesignerAccess != "unknow") {
			syra_site.designStore.showPageOpener(page.mainPageDesignerAccess != "disabled" && syra_site.userProfile.hasDesignRight());
		}
	}
};

function _getTopParentPage(page) {
	if (page.vignetteField) {
		page = _getTopParentPage(page.vignetteField.page);
	}
	if (page.inlinePageHost) {
		page = _getTopParentPage(page.inlinePageHost.record.page);
	}
	if (page.isLandingPage && syra_site.landingPageMaster) {
		page = syra_site.landingPageMaster;
	}
	return page;
}

function _applyMaximizeToBar(bar, maximized) {
	if (bar) {
		if (maximized) {
			bar.onBeforeMaximized = bar.$isCollapsed;
			bar.$isCollapsed = true;
			bar.ensureState();
		} else {
			if (bar.onBeforeMaximized !== undefined) {
				bar.$isCollapsed = bar.onBeforeMaximized;
				bar.ensureState();
			}
		}
	}
}

exports.maximizeItem = function(item, picker) {
	var page = _getTopParentPage(item.page);
	var slot = page.scrollview;
	item.isMaximized = !item.isMaximized;
	if (item.isMaximized) {
		item._parentChildren = [];
		var nodes = slot.children;
		while (nodes.length) {
			var node = nodes[0];
			node.parentNode && node.parentNode.removeChild(node);
			item._parentChildren.push(node);
		}
		slot.appendChild(item.domItem);
	} else {
		item.layoutSlot.appendChild(item.domItem);
		syra_site.dom.empty(slot);
		for (var ii = 0, jj = item._parentChildren.length; ii < jj; ii++) {
			slot.appendChild(item._parentChildren[ii]);
		}
		delete item._parentChildren;
	}
	_applyMaximizeToBar(page.menuBar, item.isMaximized);
	if (page.fusionBar) {
		item.$item && item.$item.$isNavigationList && page.fusionBar.toggleBar(!item.isMaximized);
		_applyMaximizeToBar(page.fusionBar, item.isMaximized);
	}
	if (syra_site.landingPageMaster) {
		_applyMaximizeToBar(syra_site.landingPageMaster.bar, item.isMaximized);
		syra_site.landingPageMaster.resizeArticle(true);
	} else {
		item.page.resizeArticle(true);
	}

	var state = item.isMaximized ? "minimize" : "maximize";
	syra_menus.updateButtonIcon(picker || item.maximizePicker, syra_local["box_" + state], state);
};

exports.getDocumentIcon = function(mediaType) {
	mediaType = mediaType.toLowerCase();
	if (mediaType.indexOf("image") >= 0) {
		return "image";
	}
	if (mediaType.indexOf("pdf") >= 0) {
		return "pdf";
	}
	if (mediaType.indexOf("excel") >= 0 || mediaType.indexOf("officedocument.spreadsheet") >= 0) {
		return "excel";
	}
	if (mediaType.indexOf("mailmerge") >= 0) {
		return "mailmerge";
	}
	if (mediaType.indexOf("word") >= 0) {
		return "msword";
	}
	if (mediaType.indexOf("ms-powerpoint") >= 0 || mediaType.indexOf("officedocument.presentationml") >= 0 || mediaType.indexOf("ppt") >= 0) {
		return "powerpoint";
	}
	if (mediaType.indexOf("xml") >= 0) {
		return "xml";
	}
	if (mediaType.indexOf("html") >= 0) {
		return "html";
	}
	if (mediaType.indexOf("json") >= 0) {
		return "json";
	}
	if (mediaType.indexOf("text") >= 0) {
		return "text";
	}
	mediaType = mediaType.replace("application/", "");
	return (mediaType.indexOf("image") >= 0) ? "image" : mediaType.replace("/", "-");
};


exports.dispose = function() {
	if (syra_site) {
		syra_site.bookmarksManager && syra_site.bookmarksManager.dispose();
		syra_site.jobsPage && syra_site.jobsPage.dispose();
		_dragDropManager.dispose();
		syra_site.jobsPage = syra_site.bookmarksManager = window.syra_dd = null;
	}
};
