// syracuse-ui/lib/site/tools/siteFunctions.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var bookmarksManager = require("syracuse-ui/lib/site/bookmarks/bookmarksManager");
var LegalPage = require('syracuse-ui/lib/site/legal/legalPage').LegalPage;
var dragDropManager = require("syracuse-ui/lib/site/tools/dragDropManager");
var TopPanel = require("syracuse-ui/lib/site/tools/topPanel").TopPanel;
var spyer = require('syracuse-ui/lib/site/spy/spyer');

var _helpPopupSlot, _helpPopup;
exports.openHelpPopup = function(menuItem) {
	var site = syra_site;
	if (!_helpPopupSlot) {
		_helpPopupSlot = document.createElement("div");
		_helpPopupSlot.className = "s-site-help-links-slot";
		var $links = site.$prototype.$helpLinks;
		for (var ii = 0, jj = $links.length; ii < jj; ii++) {
			site.loadNewItem(_helpPopupSlot, {
				$bind: $links[ii],
				$category: "link",
				$skin: "s-site-help-item-link"
			});
		}
	}
	if (!_helpPopup) {
		_helpPopup = site.dialogManager.openPopup(site, {
			content: site,
			slot: _helpPopupSlot,
			position: {
				my: "left top",
				at: "left bottom",
				of: $(menuItem.domItem)
			},
			onClose: function() {
				setTimeout(function() {
					_helpPopup = null;
				}, 200);
			}
		});
	} else {
		_helpPopup.close();
	}
};

exports.openLegalPage = function(menuItem) {
	syra_controller.sendRequest(null, {
		method: "GET",
		$location: {
			$url: syra_site.$item.$legalPrototypeUrl
		}
	}, function(data, response, $url) {
		var page = new LegalPage();
		page.loadBox(data);
		if (page.menuBar) {
			page.menuBar.toggleBar(false);
		}
		syra_controller.sendRequest(null, {
			method: "GET",
			$location: {
				$url: syra_site.$item.$legalDataUrl
			}
		}, function(data, response, $url) {
			page.applyChange(data);
			syra_site.dialogManager.openModal(menuItem.page, {
				page: page
			});
		}, function(error) {
			var $diagnoses;
			if (error.data.indexOf("$diagnoses") != -1) {
				$diagnoses = JSON.parse(error.data).$diagnoses;
			} else {
				$diagnoses = [{
					$severity: "error",
					$message: error.data
				}];
			}
			syra_site.showDiagnoses({
				$diagnoses: $diagnoses
			});
		});

	}, function(error) {
		var $diagnoses;
		if (error.data.indexOf("$diagnoses") != -1) {
			$diagnoses = JSON.parse(error.data).$diagnoses;
		} else {
			$diagnoses = [{
				$severity: "error",
				$message: error.data
			}];
		}
		syra_site.showDiagnoses({
			$diagnoses: $diagnoses
		});
	});

};

exports.addBookmarkManager = function(site) {
	var bar = document.createElement("div");
	bar.id = "s-site-header-bottom-bookmarks";
	site.bookmarksManager = bookmarksManager;
	site.bookmarksManager.load(site.headerBottom.appendChild(bar));
};

exports.addJobsOpener = function(site, slot) {
	var picker = site._jobsViewerOpener = document.createElement("a");
	picker.syraSiteObserver = "siteFunctions";
	picker.className = "s-jobs-viewer-opener";
	picker.syraOnClick = "onJobsOpenerClick";
	picker.setAttribute("href", "#");
	picker.style.display = "none";
	slot.appendChild(picker);
};
exports.onJobsOpenerClick = function() {
	syra_site.jobsViewer && syra_site.jobsViewer.toggle();
};

exports.addLogo = function(site, slot) {
	site.headerLogo = document.createElement("a");
	site.headerLogo.id = "s-site-header-logo";
	site.headerLogo.syraSiteObserver = "siteFunctions";
	site.headerLogo.syraOnClick = "onLogoClick";
	slot.appendChild(site.headerLogo);
};
exports.onLogoClick = function() {
	syra_site.gotoHome();
};

exports.addLogoutButton = function(site) {
	var picker = site._logoutPicker = document.createElement("a");
	picker.syraSiteObserver = "siteFunctions";
	picker.className = "s-site-logout";
	picker.syraOnClick = "onLogoutClick";
	picker.setAttribute("href", "#");
	picker.setAttribute("title", site.localize.siteLogoutLink);
	site.headerTop.appendChild(picker);
};
exports.onLogoutClick = function() {
	syra_site.logout();
};
exports.addPageDesignerOpener = function(site) {
	var picker = site.pageDesignerOpener = document.createElement("a");
	picker.syraSiteObserver = "siteFunctions";
	picker.className = "s-page-designer-opener";
	picker.syraOnClick = "onPageDesignerClick";
	picker.style.visibility = "hidden";
	picker.setAttribute("href", "#");
	site.headerTop.appendChild(picker);
};
exports.onPageDesignerClick = function() {
	var site = syra_site;
	var page;
	exports.toggleTopPanel(false);
	if (site.mainPage.isFusionPage) {
		page = site.fusionGateway.activatedBook.selectedSheet;
	} else {
		page = site.dialogManager.getTopDialogPage();
		if (page) {
			page = page._content;
		}
	}
	page = page || site.mainPage;
	if (page) {
		page.designArticle(true);
		page.resizePage(true);
	}
};

exports.addDragDropManager = function(site) {
	window.syra_dd = dragDropManager;
};

exports.loadSpyer = function(site) {
	site.spyer = spyer;
};


var _topPanel;

exports.appendTopPanelOpener = function(layoutSlot) {
	var opener = document.createElement("a");
	opener.syraSiteObserver = "siteFunctions";
	opener.className = "s-site-top-pn-opener-link";
	opener.syraOnClick = "onTopPanelClick";
	opener.setAttribute("href", "#");
	_topPanel = {
		opener: opener
	};
	layoutSlot.appendChild(opener);
};
exports.onTopPanelClick = function() {
	exports.toggleTopPanel(true);
};
exports.toggleTopPanel = function(show, event, onBeforeClick) {
	if (_topPanel) {
		var site = syra_site;
		var isVisible;
		if (show) {
			if (!_topPanel.panel) {
				_topPanel.container = document.createElement("div");
				_topPanel.container.className = "s-site-top-pn";
				_topPanel.container.style.display = "none";
				site.layoutSlot.appendChild(_topPanel.container);
				_topPanel.panel = new TopPanel();
				_topPanel.panel.$prototype = site.$prototype;
				_topPanel.panel.page = site;
				_topPanel.panel.layoutSlot = _topPanel.container;
				_topPanel.panel.loadBox();
			}
			if (_topPanel.container.style.display == "none") {
				isVisible = true;
				var rect = _topPanel.opener.getBoundingClientRect();
				if (document.dir === 'rtl') {
					_topPanel.container.style.left = rect.left + "px";
				} else {
					_topPanel.container.style.right = (site.getLayoutSize().right - rect.right) + "px";
				}
				if (_topPanel.panel.navigationBar) {
					_topPanel.panel.navigationBar.domItem.style.display = (site.fusionGateway && site.fusionGateway.activatedBook) ? "none" : "";
				}
			}
		} else {
			if (onBeforeClick && event &&
				((event.target == _topPanel.opener) ||
					_topPanel.opener.contains(event.target))) {
				return;
			}
			if (_topPanel.panel) {
				var panel = _topPanel.panel;
				if (event) {
					var isNavigationLink = panel.navigationBar && panel.navigationBar.domItem.contains(event.target);
					if (!isNavigationLink && panel.domItem.contains(event.target)) {
						return;
					}
				}
			}
		}
		if (_topPanel.container) {
			_topPanel.container.style.display = isVisible ? "" : "none";
		}
		if (_topPanel.opener) {
			site.dom.toggleClass(_topPanel.opener, "s-open", isVisible);
		}
	}
};

exports.showUserIdentity = function(userProfile, newData) {
	if (_topPanel && _topPanel.opener && userProfile.dataset.user && !syra_site.nameDiv) {
		// append user profile info (firstname, lastname and photo)
		// name info
		syra_site.nameDiv = document.createElement("div");
		syra_site.nameDiv.className = "s-site-user-name";
		syra_site.nameDiv.textContent = (userProfile.dataset.user.firstName ? userProfile.dataset.user.firstName + " " : "") + (userProfile.dataset.user.lastName ? userProfile.dataset.user.lastName : "");

		_topPanel.opener.appendChild(syra_site.nameDiv);

		// photo info
		var photoDiv = document.createElement("div");
		photoDiv.className = "s-site-user-photo";
		if (userProfile.dataset.user.photo) {
			userProfile.loadNewItem(photoDiv, {
				$css: "s-site-user-photo",
				$skin: "s-site-user-photo",
				$inplace: true,
				$imageWidth: " ",
				$imageHeight: " ",
				$bind: "photo",
				$category: "field",
				$field: {
					$type: "image",
					$url: syra_site.expressionMaker.parse(userProfile, userProfile.dataset.user.photo.$url, {
						$uuid: userProfile.dataset.user.$uuid,
						$pluralType: "users",
						$key: userProfile.$prototype.$key,
						$baseUrl: userProfile.$prototype.$baseUrl
					})
				}
			});

			var newPhoto = helpers.object.clone(userProfile.dataset.user.photo);
			newPhoto.$url = syra_site.expressionMaker.parse(userProfile, userProfile.dataset.user.photo.$url, {
				$pluralType: "users",
				$key: userProfile.dataset.user.$uuid,
				$baseUrl: userProfile.$prototype.$baseUrl
			});
			newData.photo = newPhoto;
		} else {
			var mn = document.createElement("a");
			mn.className = "s-site-user-nophoto";
			photoDiv.appendChild(mn);
		}
		_topPanel.opener.appendChild(photoDiv);
	}
};




exports.dispose = function(site) {
	if (site.bookmarksManager) {
		site.bookmarksManager.dipsose();
	}
	dragDropManager.dispose();

	if (site.spyer) {
		site.spyer.dispose();
	}
	site.bookmarksManager = window.syra_dd = site.spyer = null;
	_topPanel = helpPopupSlot = _helpPopup = null;
};
