// syracuse-ui/lib/site/tools/store.js
"use strict";

var _map = {};

exports.getMap = function() {
	return _map;
};

exports.add = function(item) {
	_map[item.id] = item;
};

exports.get = function(id) {
	return _map[id];
};

exports.remove = function(item) {
	delete _map[item.id];
};

exports.findRecordByServerIndex = function(list, $serverIndex) {
	for (var ii = 0, jj = list.records.length; ii < jj; ii++) {
		if (list.records[ii].$serverIndex == $serverIndex) {
			return list.records[ii];
		}
	}
	return null;
};

exports.findRecord = function(list, dom) {
	var record;
	while (dom && dom != document) {
		if (dom.syraItem !== undefined) {
			record = _map[dom.syraItem];
			if (record && record.isRecordArticle && !record.isChildFieldRecord) {
				return record;
			}
		}
		dom = dom.parentNode;
	}
	return null;
};


exports.findArticle = function(dom) {
	while (dom != null) {
		if (dom.syraarticle) {
			return _map[dom.syraarticle];
		}
		dom = dom.parentNode;
	}
};
exports.findField = function(dom) {
	var field;
	while (dom != null) {
		if (dom.syraItem) {
			field = _map[dom.syraItem];
			if (!field || (field && field.isField)) {
				break;
			}
		}
		dom = dom.parentNode;
	}
	return field;
};

exports.findItem = function(dom) {
	var item;
	while (dom != null) {
		if (dom.syraItem) {
			item = _map[dom.syraItem];
			if (item) {
				break;
			}
		}
		dom = dom.parentNode;
	}
	return item;
};

exports.findList = function(item) {
	var list = item.isArrayField && item;
	if (!list) {
		//walk up from field
		var article = item.articleParent;
		while (article) {
			if (article.isArrayField) {
				list = article;
				break;
			}
			article = article.articleParent;
		}
	}
	if (!list && item.page.boundFields) {
		//walk up from page
		var binds = Object.keys(item.page.boundFields);
		for (var ii = 0, jj = binds.length; ii < jj; ii++) {
			var field = item.page.boundFields[binds[ii]];
			field = field && field[0];
			if (field && field.isArrayField) {
				list = field;
				break;
			}
		}
	}
	return list;
};


exports.getLists = function(page) {
	var lists = [];
	var $binds = page.boundFields ? Object.keys(page.boundFields) : [];
	for (var ii = 0, jj = $binds.length; ii < jj; ii++) {
		var fields = page.boundFields[$binds[ii]];
		if (fields) {
			for (var mm = 0, kk = fields.length; mm < kk; mm++) {
				if (fields[mm] && fields[mm].isArrayField) {
					lists.push(fields[mm]);
				}
			}
		}
	}
	return lists;
};

exports.dispose = function() {
	_map = null;
};
