// syracuse-tablet/html/js/wc/controllers/homeController.js
"use strict";

var Base = require("syracuse-tablet/html/js/wc/controllers/sdataController").SdataController;
var utils = require('syracuse-tablet/html/js/wc/helpers/utils');
var modules = require('syracuse-tablet/html/js/wc/common/modules');


exports.HomeController = utils.defineClass(
	function HomeController(dataset) {
		console.log("HomeController", dataset)
		Base.call(this, dataset);
		this._dirty = false;
	},
	Base, {
		destroy: function() {
			Base.prototype.destroy.call(this);
		},
		headerFooterToggle: function() {
			//var isSmartphone = modules.get("siteLayout").getDeviceType() === "smartphone";
			if (this._header == null) {
				var panelInfo;
				panelInfo = {
					name: "headerEditHome",
					topIndex: 0,
					selector: "header",
					$type: "application/x-panel-header-edithome-tablet"
				}
				this._header = this.uiAdapter.attachedControlsMgr.createAddtionalControl(panelInfo, {});
			}
			if (this._footer == null) {
				panelInfo = {
					name: "footerEditHome",
					topIndex: 0,
					selector: "footer",
					$type: "application/x-panel-footer-edithome-tablet"
				}
				this._footer = this.uiAdapter.attachedControlsMgr.createAddtionalControl(panelInfo, {});
			}
			this.uiAdapter.attachedControlsMgr.toggleControls("headerEditHome", "header")
			this.uiAdapter.attachedControlsMgr.toggleControls("footerEditHome", "footer")
		},
		isDirty: function() {
			return this._dirty;
		},

	}
);
