// @sage/syracuse-react/src/test/navigation/breadcrumbTest.tsx
"use strict";
import * as React from 'react';
import { IPageProps } from '../../page/page';
import Section from '../../section/section';
import * as Sdata from '../../dataStore/sdata';
import { Breadcrumb } from '../../navigation/breadcrumb/breadcrumb';
import { BreadcrumbPanel } from '../../navigation/breadcrumb/breadcrumbPanel';

export default class BreadcrumbTest extends React.Component<IPageProps, {}>{
    static representationName: string = "breadcrumb";
    get modules(): any {
        return this.props.profile.sitemap.modules;
    }

    get steps(): Sdata.IBreadcrumStep[] {
        return [{
            "title": "All"
        }, {
            "title": "Administration",
            "$uuid": "c6ffde41-c91e-4855-9fe6-ba72cdbe2ec3"
        }, {
            "title": "Administration",
            "$uuid": "7571c160-aeda-41d4-aedc-f974b815f863"
        }, {
            "title": "Settings",
            "$uuid": "45ed162d-bf3e-4800-a2f3-ae81a28b65ec"
        }];
    }
    get selectedIds(): any {
        return {
            "moduleId": "c6ffde41-c91e-4855-9fe6-ba72cdbe2ec3",
            "submoduleId": "7571c160-aeda-41d4-aedc-f974b815f863",
            "blockIds": ["45ed162d-bf3e-4800-a2f3-ae81a28b65ec"]
        };
    }

    render() {
        let module = this.steps[1];
        let submodule = this.steps[2]
        return <Breadcrumb steps={this.steps} modules={this.modules} />
        // return <div/>;
        /*return < WebPage title="Breadcrumb">
            <Section title="Breadcrumb Enabled">
                <Breadcrumb steps={this.steps} modules={this.modules}/>
            </Section>
            <Section title="Breadcrumb Disabled">
                <Breadcrumb steps={this.steps} disabled={true} modules={this.modules}/>
            </Section>
            <Section title="Breadcrumb Panel Modules">
                <BreadcrumbPanel modules={this.props.profile.sitemap.modules} selectedIds = {this.selectedIds}/>
            </Section>
            <Section title="Breadcrumb Panel SubModules">
                <BreadcrumbPanel moduleId={this.steps[1].$uuid} modules={this.props.profile.sitemap.modules} selectedIds = {this.selectedIds}/>
            </Section>
            <Section title="Breadcrumb Panel Blocks">
                <BreadcrumbPanel moduleId={module.$uuid} submoduleId={submodule.$uuid} modules={this.props.profile.sitemap.modules} selectedIds = {this.selectedIds}/>
            </Section>
            <Section title="Breadcrumb Panel Subblocks">
                <BreadcrumbPanel moduleId={module.$uuid} submoduleId={submodule.$uuid} blockIds={[this.steps[3].$uuid]} modules={this.props.profile.sitemap.modules} selectedIds = {this.selectedIds}/>
            </Section>
        </ WebPage>;*/
    }
}
