// syracuse-ui/lib/office/word/wordSite.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var siteHelper = require("syracuse-ui/lib/common/site/siteHelper");
var DesktopSite = require('syracuse-ui/lib/desktop/site/desktopSite').DesktopSite;

var Controller = require("./controller").Controller;

function WordSite(){
}

exports.WordSite = helpers.defineClass(WordSite, DesktopSite, {
    drawBox: function(){
        var self = this;
        self._renderHeader();
        self.$$body = $("<div id='s-site-body'/>").appendTo(self.$$container);
        $(window).bind("resize", function(evt){
            self.resize();
        });

        document.site.logon(function() {
            document.controller.startNavigation();
            self.$$container.show();
            self.resize();
            
            try {
            	self.wordInterface.onLogin();
            } catch(e) {
            }
        });
    },
    
    updateDocumentTitle: function(){
        var title = this.mainPage.getTitle();
        if (title) {
            var record = this.userProfile.dataset;
            this.mainPage.drawUserProfile(record);
            
            document.title = title;
        }
    },
    _renderHeader: function(){
        this.$$userProfileHost = $("<div id='s-user-profile'/>").appendTo(this.$$container);
    },
    _renderSearch: function($$container){
    }
});

exports.load = function($item, $prototype) {
	var self = this;
	
    var widgetsLibrary = require('syracuse-ui/lib/desktop/site/widgetsLibrary');
    widgetsLibrary.pageCategories.wordmailmerge = require("./wordPage").WordPage;
    var site = siteHelper.loadSite({
        siteClass: WordSite,
        widgetsLibrary: widgetsLibrary,
        messageBoxClass: require('syracuse-ui/lib/desktop/site/messageBox').MessageBox,
        diagnoseViewerClass: require('syracuse-ui/lib/desktop/site/diagnoseViewer').DesktopDiagnoses,
        userProfileClass: require("./userProfile").UserProfile,
        controllerClass: Controller,
        $item: $item,
        $prototype: $prototype
    });

    site.agents = {
    		createMailMergeDS: {
            saveDSWord: function(article, menu, record) {
            	site.wordInterface.attachResourceUrl(record.serviceUrl);
            	site.wordInterface.onLogin();
            },
        }
    }

    var wordInterface = require('msoffice/lib/word/wordInterface').WordInterface;
    site.wordInterface = new wordInterface();
    site.$syracuseMainPageUrl = $item.$syracuseMainPageUrl;
    return site;
};
