// syracuse-ui/lib/office/word/wordSaveSite.js
"use strict";

// This site is only used to exchange the controller an catch the save event of a document entry to directly
// save content from an active document inside ms word

var helpers = require('syracuse-core/lib/helpers');
var Site = require('syracuse-ui/lib/site/site').Site;
var OfficePage = require('syracuse-ui/lib/office/officePage').OfficePage;
var Controller = require("./controller").Controller;

function WordSafeSite() {}

exports.WordSafeSite = helpers.defineClass(WordSafeSite, Site, {
	_onLogon: function() {
		if (!this._isLocalChanging) {
			this.isUserLogonEnd = true;
			document.site.history.start();
			try {
				// Access the current document url (URL where the doc is saved in storage area)
				// if the url  is set, directly safe w/o showing edit page
				var docUrl = document.site.wordInterface.getDocumentUrl();
				if (docUrl && docUrl !== "") {
					if (/syracuse\/collaboration\/syracuse/.test(docUrl)) {
						document.site.wordInterface.publishDocument();
					} else {
						document.site.wordInterface.publishDocumentX3(docUrl);
					}
				}
			} catch (e) {}
			this.layoutSlot.style.display = "";
			this.resize();
		}
	},
	_renderHeader: function() {

	}
});

exports.load = function($item, $prototype) {
	var widgetsLibrary = require('syracuse-ui/lib/site/widgetsLibrary');
	widgetsLibrary.defaultPageCategory = OfficePage;
	$item.$isSecurityTitleVisible = false;
	var site = (new WordSafeSite()).loadBox({
		isUserEndpointChangeDisabled: true,
		widgetsLibrary: widgetsLibrary,
		controllerClass: Controller,
		$item: $item,
		$prototype: $prototype
	});

	var wordInterface = require('msoffice/lib/word/wordInterface').WordInterface;
	site.wordInterface = new wordInterface();
	site.$syracuseMainPageUrl = $item.$syracuseMainPageUrl;
	return site;
};
