// syracuse-ui/lib/office/word/wordSaveSite.js
"use strict";

// This site is only used to exchange the controller an catch the save event of a document entry to directly
// save content from an active document inside ms word

var helpers = require('syracuse-core/lib/helpers');
var Site = require('syracuse-ui/lib/site/site').Site;
var OfficePage = require('syracuse-ui/lib/office/officePage').OfficePage;
var Controller = require("./controller").Controller;

function WordSafeSite(){
}

exports.WordSafeSite = helpers.defineClass(WordSafeSite, Site, {
    drawBox: function(){
        var self = this;
        self._renderHeader();
        self.$$body = $("<div id='s-site-body'/>").appendTo(self.$$container);
        self.logon(function(){
            document.controller.startNavigation();
            self.$$container.show();
            self.resize();
            
            try {
                // Access the current document url (URL where the doc is saved in storage area)
                // if the url  is set, directly safe w/o showing edit page
                var docUrl = document.site.wordInterface.getDocumentUrl();
                if (docUrl && docUrl !== "") 
                    document.site.wordInterface.publishDocument();
            } 
            catch (e) {
            }
        });
    },
    updateDocumentTitle: function(){
        var title = this.mainPage.getTitle();
        if (title) {
            var record = this.userProfile.dataset;
            this.mainPage.drawUserProfile(record);
            
            document.title = title;
        }
    },
    _renderHeader: function(){
        this.$$userProfileHost = $("<div id='s-user-profile'/>").appendTo(this.$$container);
    },
    _renderSearch: function($$container){
    }
});

exports.load = function($item, $prototype){
    var self = this;
    
    var widgetsLibrary = require('syracuse-ui/lib/site/widgetsLibrary');
    widgetsLibrary.defaultPageCategory = OfficePage;
    var site = (new WordSafeSite()).loadBox({
        widgetsLibrary: widgetsLibrary,
        userProfileClass: require("syracuse-ui/lib/office/userProfile").UserProfile,
        controllerClass: Controller,
        $item: $item,
        $prototype: $prototype
    });
    
    var wordInterface = require('msoffice/lib/word/wordInterface').WordInterface;
    site.wordInterface = new wordInterface();
    site.$syracuseMainPageUrl = $item.$syracuseMainPageUrl;
    return site;
};
