// syracuse-ui/lib/office/word/wordSite.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Site = require('syracuse-ui/lib/site/site').Site;
var Controller = require("syracuse-ui/lib/controller/controller").Controller;
var OfficePage = require('syracuse-ui/lib/office/officePage').OfficePage;

function WordSite(){
}

exports.WordSite = helpers.defineClass(WordSite, Site, {
    drawBox: function(){
        var self = this;
        self._renderHeader();
        self.$$body = $("<div id='s-site-body'/>").appendTo(self.$$container);
        self.logon(function(){
            document.controller.startNavigation();
            self.$$container.show();
            self.resize();
            
            try {
                self.wordInterface.onLogin();
            } 
            catch (e) {
            }
            
            var addinVersion = self.wordInterface.getAddinVersion();
        });
    },
    
    updateDocumentTitle: function(){
        var title = this.mainPage.getTitle();
        if (title) {
            var record = this.userProfile.dataset;
            this.mainPage.drawUserProfile(record);
            
            document.title = title;
        }
    },
    _renderHeader: function(){
        this.$$userProfileHost = $("<div id='s-user-profile'/>").appendTo(this.$$container);
    },
    _renderSearch: function($$container){
    }
});

exports.load = function($item, $prototype){
    var self = this;
    
    var widgetsLibrary = require('syracuse-ui/lib/site/widgetsLibrary');
    widgetsLibrary.defaultPageCategory = OfficePage;
    widgetsLibrary.pageCategories.wordmailmerge = require("./wordPage").WordPage;
    widgetsLibrary.pageCategories.wordreport = require("./wordReportPage").WordReportPage;
    var site = (new WordSite()).loadBox({
        widgetsLibrary: widgetsLibrary,
        userProfileClass: require("syracuse-ui/lib/office/userProfile").UserProfile,
        controllerClass: Controller,
        $item: $item,
        $prototype: $prototype
    });
    
    site.agents = {
        createMailMergeDS: {
            saveDSWord: function(article, menu, record){
                site.wordInterface.attachResourceUrl(record.serviceUrl);
                site.wordInterface.onLogin();
            },
        }
    }
    
    var wordInterface = require('msoffice/lib/word/wordInterface').WordInterface;
    site.wordInterface = new wordInterface();
    site.$syracuseMainPageUrl = $item.$syracuseMainPageUrl;
    return site;
};
