// syracuse-ui/lib/office/word/wordReportPage.js
"use strict"

var helpers = require('syracuse-core/lib/helpers');
var page = require('syracuse-ui/lib/common/article/page');
var article = require('syracuse-ui/lib/common/article/article');

function WordReportPage() {
}

exports.WordReportPage = helpers.defineClass(WordReportPage, page.Page, {
    drawBox: function() {
    	var self = this;
    	self.wordLayoutBoxes = new Array();
    	self.createLayout(self.$item, self.$prototype.$, null, self.wordLayoutBoxes, null, self.$prototype);
    },
    createWordTemplate: function(wordInterface) {
    	var self = this;
    	wordInterface.createWordTemplate(self.wordLayoutBoxes);
    },
    populateWordTemplate: function(wordInterface) {
    	var self = this;
    	wordInterface.populateWordTemplate(self.createSimpleData());
    },
    onDialogResize: function() {
    },
    applyChange: function(newData, response, requestUrl) {
    	var self = this;
    	self.pageData = newData;
    },
    dispose: function() {
    },
    createLayout: function(article, prototype, parent, boxes, level, baseProto) {
    	// walk layout and build flat list of boxes which contain properties with type information
    	var self = this;
    	
    	var layoutType;
    	var bind;
    	var layoutSubType;
    	var data;
    	var box;
    	var hidden;
    	var i;

    	function getLocalizedText(val) {
    		if (val === "{$title}")
    			return baseProto.$title || val;
    		return val && val.replace(/\{(@\w+)\}/g, function(match, p1) {
    			return (baseProto.$localization && baseProto.$localization[p1]) || match;
    		});
    	}

    	if (!level)
    		level = 1;

    	var items = article.$items || (article.$layout && article.$layout.$items);
    	
    	if (article.$bind && items) {
    		// collection
    		bind = article.$bind; // collection property
    		data = (prototype && bind) ? prototype[bind] : prototype;
    		if (!data || data.$type.indexOf("application/x-collection") < 0)
    		{
    			// Not handled yet
    			console.log("unknown type: " + (data && data.$type) + "(" + bind + ")");
    		}
    		else
    		{
	    		data = data.$item.$;
	    		box = {
	    			$title:  getLocalizedText(article.$title),
	    			$container: "table",
	    			$bind: bind
	    		};
    		}
    	} else if (article.$bind && !items) {
    		// property
    		bind = article.$bind; // property
    		hidden = article.$isHidden || false
    		data = (prototype && bind) ? prototype[bind] : prototype;
    		if (data) {
    			parent.$items = parent.$items || {};
    			parent.$items[bind] = {
    					$type: data.$type,
    					$title: data.$title || bind,
    					$bind: bind,
    					$hidden: hidden
    			};
    		}
    	} else {
    		// container
    		data = prototype;
    		box = {
    				$title:  getLocalizedText(article.$title),
    				$container: "box"
    		};
    	}
    	
    	if (box && box.$title) { // a box with a title is considered as one level of indentation of headers
    		box.$level = level;
    		level ++;
    	}

    	var childBoxes = new Array();
    	items && items.map(function(item) {
    		self.createLayout(item, data, box, childBoxes, level, baseProto);
    	});
    	if (box) {
    		if (box.$title || box.$items)
    			boxes.push(box);
    	}
    	for (i = 0; i < childBoxes.length; i++) {
    		boxes.push(childBoxes[i]);
    	}
    },
    createSimpleData: function() {
    	var self = this;
    	
    	function addProperties(proto, dataOut, dataIn) {
    		for (var property in proto)
    		{
    			var p	 = proto[property];
    			var prop = {};
    			dataOut[property] = prop;
    			
    			prop.$type = p.$type;
    			if (prop.$type === "application/x-collection") {
    				var dataArray = dataIn[property];
    				if (dataArray) {
	    				prop.$items = new Array();
	    				for (var i = 0; i < dataArray.length; i++) {
	    					var item = {};
	    					prop.$items.push(item);	
	    					addProperties(p.$item.$, item, dataArray[i]);
	    				}
    				}
				}
    			else if (prop.$type === "application/x-reference") {
    				addProperties(p.$, prop, p);
    			}
    			else
   				{
    				// Todo: Review / Check for Lists!
    				var result;
    				var value = dataIn[property];
    				if (value && value.$url) {
    					result = {};
    					result.$url = self.parseExpression(value.$url, self.pageData);
    				} else {
    					result = value;
    				}
    				prop.$value = result;
   				}
    		}
    	};

    	var dataOut = {};
    	addProperties(self.$prototype.$, dataOut, self.pageData);
    	return dataOut;
    },
    setLineCount: function(lineCount) {
    }
});

// Called from word when additional data has to be fetched (eg. images)
window.readURLContent = function(url) {
	alert(url);
	/*
	var result = $.ajax({
        url: result.$url,
        type: 'GET',
        async: false,
        cache: false,
        timeout: 30000,
        error: function() {
        	alert("error");
            return true;
        },
        success: function(data) {
        	alert(data);
        }
    });
	alert(result);
	*/
	return result;
};
