// syracuse-ui/lib/field/array/gridBuilder/card/cardDecorator.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function _renderRecordCard(record, slot) {
	slot.setAttribute("data-s-record", slot.syraRecord = record.$uuid);
	record.setArticleId(slot);
	if (!record.cardItem) {
		record.$item = record.list.$item.$cardItem;
		record.$skin = "s-grid-carditem";
		record.layoutSlot = slot;
		record.dataSlot = record.domItem = record.body = document.createElement("div");
		record.layoutSlot.appendChild(record.domItem);
		record.ensureLayoutMode();
		record.renderLayoutContent();
		record.cardItem = record;
	}
	slot.appendChild(record.body);
}


function CardDecorator() {}

exports.CardDecorator = helpers.defineClass(CardDecorator, null, {
	removeRecordCards: function() {
		var list = this.list;
		var col = list.builder.columnsMap.rowCard;
		if (col) {
			list.builder.allColumns.splice(list.builder.allColumns.indexOf(col), 1);
			col.table.columns.splice(col.table.columns.indexOf(col), 1);
			syra_site.dom.removeChild(col.titleCell);
			var record = list.builder.filterRecord;
			if (record) {
				syra_site.dom.removeChild(record.cellsMap[col.key]);
				delete record.cellsMap[col.key];
			}
			for (var ii = 0, jj = list.records.length; ii < jj; ii++) {
				var record = list.records[ii];
				record.rowCardOpener && syra_site.dom.removeChild(record.rowCardOpener.parentNode);
				if (record.layoutContent) {
					syra_site.layoutUpdater.clearContent(record.layoutContent);
					record.layoutContent.dispose();
					record.rowCard && syra_site.dom.removeChild(record.rowCard.row);
					list.builder.disposeRecordCard(record);
					delete record.cardItem;
				}
				delete record.rowCardOpener;
				delete record.cellsMap[col.key];
			}
			delete list.builder.columnsMap.rowCard;
			this.$rowFieldBinds = null;
		}
	},

	load: function(list) {
		this.list = list;
		if (this.list.$item.$inlineCard || this.list.$item.$cards) {
			this._convertOldCardItem();
		}
		this.appendRowCard();
		this._ensureMode();
	},
	setRowCardOpenerIcon: function(opener, expanded) {
		opener.syraExpanded = expanded;
		opener.innerHTML = expanded ? syra_menus.fontIcons.collapse_row : syra_menus.fontIcons.expand_row;
	},
	_ensureMode: function() {
		this.isOutCardMode = this.isRowCardMode = false;
		if (this.list.$item.$cardItem) {
			this.isOutCardMode = !(this.isRowCardMode = this.list.$item.$cardItem.$position == "row");
		}
	},
	onShowRecordSelection: function(record, selected) {
		if (this.isRowCardMode && record.rowCard) {
			syra_site.dom.toggleClass(record.rowCard.row, "s-list-record-selected", selected);
		}
		if (selected) {
			var popupCardBuilder = this.list.builder.popupCardBuilder;
			popupCardBuilder && popupCardBuilder.onShowRecordSelection(record, true);
			if (this.isOutCardMode) {
				syra_site.dom.empty(this.outCardSlot);
				_renderRecordCard(record, this.outCardSlot);
			}
		}
	},
	toggleRowCard: function(record, show) {
		if (record) {
			this.setRowCardOpenerIcon(record.rowCardOpener, show);
			if (!record.rowCard) {
				record.rowCard = record.list.builder.createCardRow(record);
				_renderRecordCard(record, record.rowCard.contentSlot);
			}
			record.rowCard.row.style.display = show ? "" : "none";
		}
	},
	appendRowCardOpener: function(record) {
		var th = document.createElement("td");
		th.className = this.list.cssCell + " s-grid-rowcard-cell-picker";
		record.rowCardOpener = syra_menus.addIconButton(syra_local.flShowCard, "s-grid-rowcard-picker s-fonticon-btn", "onRowCardClick");
		this.setRowCardOpenerIcon(record.rowCardOpener, false);
		th.appendChild(record.rowCardOpener);
		return th;
	},
	onRowCardOpenerClick: function(opener, record) {
		var show = !opener.syraExpanded;
		var records = record ? [record] : this.list.records;
		for (var ii = 0, jj = records.length; ii < jj; ii++) {
			this.toggleRowCard(records[ii], show);
			show && records[ii].ensureArticleVisibility(true);
		}
		var title;
		if (record) {
			title = show ? syra_local.flCloseCard : syra_local.flShowCard;
		} else {
			title = show ? syra_local.flCloseAllCards : syra_local.flShowAllCards;
		}
		syra_menus.setButtonTitle(opener, title);
		this.setRowCardOpenerIcon(opener, show);
	},
	appendRowCard: function(designing) {
		if (!this.list.$item.$cardItem) {
			if (this.list.$complexBinds.length) {
				this.list.$item.$cardItem = {
					$position: "row",
					$layout: {
						$layoutType: "tabs",
						$items: []
					}
				};
				for (var ii = 0, jj = this.list.$complexBinds.length; ii < jj; ii++) {
					var $bind = this.list.$complexBinds[ii];
					var $field = this.list.$fields[$bind];
					var $item = {
						$bind: $bind,
						$isTitleHidden: true
					};
					// adding $item properties (necessary for job viewer diagnose stacktrace view)
					if ($field.$item) {
						if ($field.$item.$isEditMode != undefined) {
							$item.$isEditMode = $field.$item.$isEditMode;
						}
						if ($field.$item.$isTitleRowHidden != undefined) {
							$item.$isTitleRowHidden = $field.$item.$isTitleRowHidden;
						}
						if ($field.$item.$isPagerHidden != undefined) {
							$item.$isPagerHidden = $field.$item.$isPagerHidden;
						}
						if ($field.$item.$isMenuRecordHidden != undefined) {
							$item.$isMenuRecordHidden = $field.$item.$isMenuRecordHidden;
						}
					}
					this.list.$item.$cardItem.$layout.$items.push({
						$category: "section",
						$title: $field.$shortTitle || $field.$title || ("( " + $bind + " )"),
						$layout: {
							$items: [$item]
						}
					});
				}
			} else {
				if (designing) {
					this.list.$item.$cardItem = (this.list.$designing && this.list.$designing.grid && this.list.$designing.grid.$cardItem) || this.list.defineDefaultCard();
				}
			}
		}
		if (this.list.$item.$cardItem) {
			this.$rowFieldBinds = syra_page.getDefinedFieldBinds(this.list.$item.$cardItem.$layout ? this.list.$item.$cardItem.$layout.$items : this.list.$item.$cardItem.$items);
		}
	},
	onRecordApplyChange: function(record) {
		if (!this.list.$isEditMode && record.dataset && record.rowCardOpener) {
			var hasData;
			for (var ii = 0, jj = this.$rowFieldBinds.length; ii < jj; ii++) {
				var data = record.dataset[this.$rowFieldBinds[ii]];
				if (data !== null && data !== undefined) {
					if (data.length !== undefined && !data.length) {
						continue;
					}
					hasData = true;
					break;
				}
			}
			record.rowCardOpener.style.visibility = hasData ? "visible" : "hidden";
		}
	},

	applyDesignMetaData: function(metaData, designing) {
		var $cardItem = this.list.$item.$cardItem && helpers.object.clone(this.list.$item.$cardItem, true);
		if (metaData.$cardPosition == "no" || metaData.$cardPosition == "row") {
			this.removeRecordCards();
			delete this.list.$item.$cardItem;
			this.list.appendCardGraphSlots();
			if (metaData.$cardPosition == "row") {
				var builder = this.list.builder;
				this.list.$item.$cardItem = $cardItem;
				if (!this.list.$item.$cardItem) {
					this.list.$item.$cardItem = (this.list.$designing && this.list.$designing.grid && this.list.$designing.grid.$cardItem) || this.list.defineDefaultCard();
				}
				this.list.$item.$cardItem.$position = "row"; //ensure position
				this.appendRowCard();
				this._ensureMode();

				var col = (builder.freezeTable && builder.freezeTable.columns.length ? builder.freezeTable : builder.scrollTable).addRowCardCol();
				var row = col.titleCell.parentNode;
				if (col.$rowIndex) {
					builder.allColumns.splice(1, 0, builder.allColumns.pop());
					col.table.columns.splice(1, 0, col.table.columns.pop());
					row.insertBefore(col.titleCell, col.$rowIndex.titleCell.nextSibling);
				} else {
					builder.allColumns.unshift(builder.allColumns.pop());
					col.table.columns.unshift(col.table.columns.pop());
					row.insertBefore(col.titleCell, row.firstChild);
				}
				var record = builder.filterRecord;
				if (record) {
					var row = record[col.table.$rowKey];
					row.insertBefore(record.cellsMap[col.key], record.cellsMap.$rowIndex ? record.cellsMap.$rowIndex.nextSibling : row.firstChild);
				}
				for (var ii = 0, jj = this.list.records.length; ii < jj; ii++) {
					var record = this.list.records[ii];
					var row = record[col.table.$rowKey];
					row.insertBefore(this.appendRowCardOpener(record), record.cellsMap.$rowIndex ? record.cellsMap.$rowIndex.nextSibling : row.firstChild);
					if (ii == 0) {
						this.toggleRowCard(record, true);
					}
				}
				builder.resizeScrollView();
			} else {
				this._ensureMode();
			}
		} else {
			this.isRowCardMode && this.removeRecordCards();

			this.list.$item.$cardItem = $cardItem;
			if (!this.list.$item.$cardItem) {
				this.list.$item.$cardItem = (this.list.$designing && this.list.$designing.grid && this.list.$designing.grid.$cardItem) || this.list.defineDefaultCard();
			}
			this.list.$item.$cardItem.$position = metaData.$cardPosition;
			this._ensureMode();
			this.list.appendCardGraphSlots();
			if (!this.list.selector || !this.list.selector.list) {
				if (this.list.$item.$selectMode === undefined) {
					this.list.$item.$selectMode = "row";
				}
				this.list.loadSelector();
			}
			if (this.list.records.length > 0) {
				var selectedRecord;
				if (this.list.selector.uuid !== undefined) {
					selectedRecord = this.list.recordsMap[this.list.selector.uuid];
				}
				if (!selectedRecord) {
					this.list.selector.selectRecord(this.list.records[0].$uuid, true);
				} else {
					this.list.showRecordSelection(selectedRecord, true);
				}
			}
		}
	},
	_convertOldCardItem: function() {
		if (this.list.$item.$inlineCard) {
			this.list.$item.$cardItem = this.list.$item.$inlineCard;
			this.list.$item.$cardItem.$position = "row";
			delete this.list.$item.$inlineCard;
		}
		if (this.list.$item.$cards) {
			switch (this.list.$item.$cards.$position) {
				case "popup":
					break;
				case "inline":
					if (!(this.list.$item.$cards.$layout && this.list.$item.$cards.$layout.$layoutType == "tabs")) {
						this.list.$item.$cardItem = this.list.$item.$cards;
					}
					if (this.list.$item.$cardItem) {
						this.list.$item.$cardItem.$position = "row";
					}
					break;
				default:
					this.list.$item.$cardItem = this.list.$item.$cards;
					if (this.list.$item.$selectMode === undefined) {
						this.list.$item.$selectMode = "row";
					}
					break;
			}
			delete this.list.$item.$cards;
		}
	},
	appendOutCardSlot: function() {
		if (!this.outCardSeparator) {
			this.outCardSeparator = document.createElement("div");
			this.outCardSeparator.syraOutCardResizer = this.list.id;
			this.outCardSeparator.syraDragSpot = this.list.id;
		}
		if (!this.outCardSlot) {
			this.outCardSlot = document.createElement("div");
		}
		var cssCell = "";
		var tableSlot = this.list.tableSlot;
		switch (this.list.$item.$cardItem.$position) {
			case "left":
				tableSlot.parentNode.insertBefore(this.outCardSeparator, tableSlot);
				tableSlot.parentNode.insertBefore(this.outCardSlot, this.outCardSeparator);
				cssCell = "-cell";
				if (this.list.$item.$cardItem.$width) {
					this.outCardSlot.style.width = this.list.$item.$cardItem.$width;
				}
				break;
			case "right":
				tableSlot.parentNode.insertBefore(this.outCardSeparator, tableSlot.nextSibling);
				tableSlot.parentNode.appendChild(this.outCardSlot, this.outCardSeparator.nextSibling);
				cssCell = "-cell";
				if (this.list.$item.$cardItem.$width) {
					this.outCardSlot.style.width = this.list.$item.$cardItem.$width;
				}
				break;
			case "top":
				var sibling = this.list.tableRow || tableSlot;
				sibling.parentNode.insertBefore(this.outCardSeparator, tableSlot);
				sibling.parentNode.insertBefore(this.outCardSlot, this.outCardSeparator);
				this.outCardSlot.style.width = "";
				break;
			case "bottom":
				var sibling = this.list.tableRow || tableSlot;
				sibling.parentNode.appendChild(this.outCardSeparator);
				sibling.parentNode.appendChild(this.outCardSlot);
				this.outCardSlot.style.width = "";
				break;
		}
		this.outCardSeparator.className = "s-list-card-sep" + cssCell;
		this.outCardSlot.className = "s-list-card-slot" + cssCell;
		tableSlot.style.display = "";
	},
	removeOutCardSlot: function() {
		syra_site.dom.removeChild(this.outCardSeparator);
		syra_site.dom.removeChild(this.outCardSlot);
		this.outCardSeparator = this.outCardSlot = null;
	},
	dispose: function() {
		this.list = this.$rowFieldBinds = this.outCardSeparator = this.outCardSlot = null;
	}
});
