// syracuse-uml/public/js/class-designer.js
"use strict";
(function(exports) {
	exports.create = function(config) {
		config = config || {};
		config.div = $(config.div || '#uml-designer');

		function initClassComponent(parent, data) {
			var $form = $('<div>').addClass('uml-form');
			$('<input>').attr('type', 'text').attr('data-prop-name', 'name').addClass('uml-field').val(data.name).appendTo($form);

			function addListWidget(name) {
				$('<hr/>').appendTo($form);
				var $div = $('<div>').attr('data-prop-name', name).appendTo($form);

				function addLine(text) {
					var $lineDiv = $('<div>');
					$('<input>').attr('type', 'text').attr('data-prop-name', 'line').addClass('uml-field').val(text).appendTo($lineDiv);
					$('<span>').text('-').click(function() {
						$lineDiv.remove();
					}).appendTo($lineDiv);
					$lineDiv.insertBefore($add);
				}
				var $add = $('<span>').text('+').click(function() {
					addLine('');
				}).appendTo($div);
				if (data[name]) data[name].forEach(addLine);
			}
			addListWidget('attributes');
			addListWidget('methods');
			$form.appendTo(parent);
		}

		function getListValue(div) {
			var values = [];
			div.find('[data-prop-name="line"]').each(function(i, input) {
				values.push($(input).val());
			});
			return values;
		}

		var $designer = config.div.addClass('uml-designer');
		var $messageDiv = $('<pre>').addClass('uml-message').appendTo($('<div>').appendTo($designer));
		var $actionsDiv = $('<div>').addClass('uml-actions').appendTo($designer);
		var $container = $('<div>').attr('id', 'uml-container').addClass('uml-container').appendTo($designer);

		var $workbench = $('<div>').attr('id', 'uml-workbench').addClass('uml-workbench').appendTo($container);

		function message(msg) {
			$designer.find('.uml-message').css({
				color: 'green'
			}).text(msg);
		}

		function error(msg) {
			return alert(msg); // for now
			$designer.find('.uml-message').css({
				color: 'red'
			}).text(msg);
		}

		function serialize(code, title) {
			var graph = {
				code: code,
				title: title,
				classes: [],
				connections: [],
			};
			$workbench.find('.uml-component').each(function(i, elt) {
				var comp = $(elt);
				var pos = comp.parent().position();
				var result = {
					id: comp.attr('data-comp-id'),
					top: pos.top,
					left: pos.left,
					name: "" + comp.find('[data-prop-name="name"]').val(),
					attributes: getListValue(comp.find('[data-prop-name="attributes"]')),
					methods: getListValue(comp.find('[data-prop-name="methods"]')),
				};
				graph.classes.push(result);
			});
			return graph;
		}

		function save() {
			try {
				//maybe find another solution to manage the position issues when the workbench is scrolled.
				$workbench.scrollLeft(0);
				$workbench.scrollTop(0);
				return serialize();
			} catch (ex) {
				error("getValue failed: " + ex.stack || ex.toString());
			}
		}

		function clear() {
			plumbInstance.detachEveryConnection();
			$workbench.empty();
		}

		// transform JSON into a graphic chart

		function deserialize(graph) {
			clear();
			var comps = [];
			graph.classes.forEach(function(clas) {
				comps[clas.id] = addComponent(clas);
			});

			plumbInstance.doWhileSuspended(function() {
				graph.connections.forEach(function(connection) {
					//determination of the anchors depending on the type of the component
					var positionAnchorsource = [];
					positionAnchorsource = ["Right", "Left"];
					var conn = plumbInstance.connect({
						source: comps[connection.source],
						target: comps[connection.target],
						type: connection.type,
						anchors: positionAnchorsource,

					});
				});
			});
			jsPlumb.fire("jsPlumbDemoLoaded", plumbInstance);
		}

		function load(graph) {
			try {
				deserialize(graph);
			} catch (ex) {
				error("Load failed: " + ex.stack || ex.toString());
			}
		}

		function createClassComponent() {
			try {
				var source = addComponent({
					connectivities: ["source"],
					top: $workbench.offset().top,
					left: $workbench.offset().left,
				});
			} catch (ex) {
				error(ex.stack || ex.toString());
			}
		}

		var plumbInstance = window.plumbInstance = jsPlumb.getInstance({
			Endpoints: [
				["Blank", {}],
				["Blank", {}]
			],
			Anchors: [
				["Right", {}],
				["Left", {}]
			],
			Connector: [
				["Flowchart"], {}
			],
			PaintStyle: {
				lineWidth: 5,
				strokeStyle: "#34B233"
			},
			ConnectionOverlays: [
				["Arrow", {
					location: 1
				}],
			],
			Container: '#uml-workbench',
		});

		plumbInstance.registerConnectionTypes({
			"basic": {
				paintStyle: {
					strokeStyle: "#34B233",
					lineWidth: 5
				},
				hoverPaintStyle: {
					lineWidth: 7
				},
				overlays: [
					["Arrow", {
						location: 1
					}],
				],
			},
		});
		plumbInstance.registerEndpointTypes({
			"source": {
				isSource: true,
				anchor: "Right",
				endpointStyle: {
					fillStyle: "#34B233"
				},
				maxConnections: -1,
			},
		});


		var lastId = 0;

		function addComponent(opts) {
			var id = ++lastId;
			var $compo = $('<div>').addClass('uml-component-outer').appendTo($workbench);

			var $compoInner = $('<div>').attr('data-comp-id', id).addClass('uml-component component').appendTo($compo).click(function(elt) {
				$compoInner.css({
					background: 'white'
				});
			});

			var $compoRemove = $('<div>').addClass('uml-remove') //
			.append('<img src ="/outputs/images/s-page-menus-open.png"/>') //
			.appendTo($compoInner).click(function(elt) {
				plumbInstance.detachAllConnections($compo);
				plumbInstance.removeAllEndpoints($compo);
				$compo.remove();
			});

			$('<div>').addClass('uml-source').appendTo($compoInner);

			// Integration of interaction buttons
			//$compoInner.append($compoRemove);
			initClassComponent($compoInner, opts);

			$compo.css({
				position: 'absolute',
				top: opts.top,
				left: opts.left
			});
			plumbInstance.draggable($compo);

			plumbInstance.makeSource($compo, {
				filter: ".uml-source",
				anchor: "Continuous",
				connector: ["Flowchart", {}],
				connectorStyle: {
					strokeStyle: "#5c96bc",
					lineWidth: 2,
					outlineColor: "transparent",
					outlineWidth: 4
				},
			});

			plumbInstance.makeTarget($compo, {
				dropOptions: {
					hoverClass: "uml-hover"
				},
				anchor: "Continuous",
				allowLoopback: true
			});

			plumbInstance.repaint($compo);

			return $compo;
		}

		var connectionStyles = {
			inheritance: {
				strokeStyle: "red",
				lineWidth: 4
			},
			association: {
				strokeStyle: "green",
				lineWidth: 4
			},
			aggregation: {
				strokeStyle: "blue",
				lineWidth: 4
			},
			composition: {
				strokeStyle: "yellow",
				lineWidth: 4
			},
		};

		function createDialog(contents, ok) {
			var $dialog = $('<div>').addClass('uml-dialog').appendTo($container).hide();
			var $outer = $('<div>').addClass('uml-dialog-outer').appendTo($dialog);
			var $inner = $('<div>').addClass('uml-dialog-inner').appendTo($outer);
			contents().appendTo($inner);
			$dialog.click(function(ev) {
				$dialog.hide();
				callback = null;
			});

			var callback;

			function open(ev, cb) {
				callback = cb;
				var r = $container[0].getBoundingClientRect();
				$dialog.css({
					'z-index': $designer.zIndex() + 100,
					position: 'absolute',
					top: $container[0].offsetTop,
					left: $container[0].offsetLeft,
					width: r.width,
					height: r.height,
					'background-color': 'transparent',
				});
				$outer.css({
					position: 'absolute',
					top: ev.clientY - r.top - 30,
					left: ev.clientX - r.left - 30,
					'background-color': 'white',
					opacity: 1,
				});
				$dialog.show();
			}

			function close(result) {
				$dialog.hide();
				callback(result);
			}
			return {
				open: open,
				close: close,
			};
		}


		var $relationDialog = createDialog(function() {
			var $ul = $('<ul>').addClass('uml-dialog-relation');
			['inheritance', 'association', 'aggregation', 'composition'].forEach(function(type) {
				$('<li>').text(type).click(function() {
					$relationDialog.close({
						relation: type,
					});
				}).appendTo($ul);
			});
			return $ul;
		});
		plumbInstance.bind("click", function(connection, ev) {
			$relationDialog.open(ev, function(result) {
				connection.setPaintStyle(connectionStyles[result.relation]);
			});
		});

		plumbInstance.bind("dblclick", function(connection) {
			plumbInstance.detach(connection);
		});

		function addButton(title, fn) {
			$('<div>').addClass('uml-button').text(title).click(function() {
				fn();
			}).appendTo($actionsDiv);
		}

		addButton('Add Class', createClassComponent);

		jsPlumb.fire("jsPlumbDemoLoaded", plumbInstance);

		return {
			getValue: save,
			setValue: load,
		};
	};
})(typeof exports !== "undefined" ? exports : registerModule('uml-designer/lib/designer'));
