// syracuse-tablet/html/js/wc/controls/panels/header/panelHeaderEditHomeTablet.js
"use strict";

var utils = require('syracuse-tablet/html/js/wc/helpers/utils');
var Base = require('syracuse-tablet/html/js/wc/controls/panels/header/panelHeaderSdataPageBase').PanelHeaderSdataPageBase;

/**
 * Dedicated header for edit homepage on tablet
 */
var _html = {
	mainContainer: '\
			<div class="s-m-header tablet s-m-dasboard-config">\
			<div class="s-m-header-title">\
				<input class="s-m-header-label" value="{{description}}">\
				{{#if template}}\
					<div class="s-m-dashboard-header-role-number">{{roleCount}}</div>\
				 <btn class="s-m-dashboard-header-btn-role fa fa-users" data-action="roleList">\
				{{/if}} \
			</div>\
			<buttons class="pull-right dashboard-header-btn" style="width:{{rightBtnWidth}}px">\
				<button type="button" class="btn btn-default" style="display:{{displaySaveButton}}" data-nevent data-naction="{{buttonSaveAction}}" data-controller-id="{{controllerId}}" data-params="{{description}}">\
					<i class="fa fa-check"></i>\
				</button>\
				<button type="button" style="float:right" class="btn btn-danger" data-nevent data-naction="closeDashboardConfig"  data-controller-id="{{controllerId}}">\
					<i class="fa fa-close"></i>\
				</button>\
			</buttons>\
		</div>'
}
var _templates = {};

function _getHtml(name, ctx) {
	var tmpl = _templates[name];
	if (!tmpl) {
		tmpl = _templates[name] = Handlebars.compile(_html[name]);
	}
	return tmpl(ctx);
};
exports.PanelHeaderEditHomeTablet = utils.defineClass(
	function PanelHeaderEditHomeTablet(controller, type, article, opts) {
		Base.call(this, controller, type, article, opts);
	}, Base, {

		destroy: function() {
			Base.prototype.destroy.call(this);
		},
		buildHtml: function() {
			Base.prototype.buildHtml.call(this);
			this.$$elmt.append(_getHtml("mainContainer", this.getHeaderCtx()));
		},
		/**
		 * Override _getItems
		 */
		getHeaderCtx: function() {
			return {
				description: this.page.getPageDescription(),
				roleCount: 0,
				displaySaveButton: this.page.controller.isDirty() ? "" : "none",
				controllerId: this.page.controller.id,
			};
		},
	});
