// syracuse-tablet/html/js/pages/actionAdapter.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var locale = require('syracuse-tablet/html/js/helpers/locale');

/**
 * Manage actions that will appear in the different panels of the page like header, footer, right panel, user profile
 * This is the base class that MUST be implemented by a child class to return action items
 * The intention of this is mostly to document and give an interface to the panel UI components
 * 
 * Most methods return a list of action items to be rendered.
 * An action item can be one of the following.
 * 
 * {
 *   "name":		  // Unique name
 * 	 "controller-id":
 *   "$action":		  // Sdata action ($save)
 *   "icon":		  // Optional, rendering control should use prototype and a default if possible
 *   "title":		  // Optional, rendering control should use prototype
 * }
 *
 * {
 *   "name":		  // Unique name
 * 	 "controller-id":
 *   "$link":   	  // Sdata link ($edit)
 *   "$bind":		  // If link is on a specific conrol and not on page level
 *   "icon":		  // Optional, rendering control should use prototype and a default if possible
 *   "title":		  // Optional, rendering control should use prototype
 * }
 * 
 * {
 *   "name":	    	// Unique name
 * 	 "controller-id":    
 * 	 "action": 			// Non sdata actions (like pin-page, represented by data-action and data-params
 *   "params":
 *   "icon":
 *   "title":
 * }
 * 
 */
exports.ActionAdapter = utils.defineClass(
	function ActionAdapter(page) {
		this.page = page;
	},
	null, {
		destroy: function() {
			this.page = null;
		},
		//
		// START: Methods to be implemented by subclasses
		//
		/**
		 * Returns items to be shown in the action panel
		 */
		getActionsPanelItems: function() {
			return [];
		},
		/**
		 * Returns items to be shown in the footer panel
		 */
		getFooterItems: function() {
			return [];
		},

		/**
		 * Returns items to be shown in the header
		 */
		getHeaderItems: function() {
			return [];
		},
		/**
		 * Returns items to be shown in the global panel
		 */
		getGlobalPanelItems: function() {
			return [];
		},
		getPageTitle: function() {
			return this.page.getPageTitle();
		},
		//
		// END: Methods to be implemented by subclasses
		//

		//
		// START: Helpers
		//
		_addDataAction: function(items, name, action, params, title, icon, controller, overrides, visibleLinks) {
			visibleLinks = visibleLinks || {}
			if (visibleLinks[name]) return;
			var item = exports.createDataAction(name, action, params, title, icon, controller, overrides);
			items.push(item);
		},

		// Standard buttons
		_addHomeButton: function(items, controller, overrides) {
			this._addDataAction(items, "home", "home", null, "global.action.title.home", "", controller, overrides);
		},
		_addPinButton: function(items, status, controller, overrides, visibleLinks) {
			status = status === "pin" ? "pin" : "unpin";
			this._addDataAction(items, "pinPage", status + "Page", null, "global.action.title." + status, "", controller, overrides, visibleLinks);
		},
		_addDesignPageButton: function(items, controller, overrides, visibleLinks) {
			this._addDataAction(items, "designPage", "designPage", null, "global.action.title.designPage", "", controller, overrides, visibleLinks);
		},
		_addShareButton: function(items, controller, overrides, visibleLinks) {
			this._addDataAction(items, "pageShare", "pageShare", null, "global.action.title.pageShare", "", controller, overrides, visibleLinks);
		},
		_addSettingsButton: function(items, controller, overrides, visibleLinks) {
			this._addDataAction(items, "settings", "settings", null, "global.action.title.settings", "settings", controller, overrides, visibleLinks);
		},
		_addAboutButton: function(items, controller, overrides, visibleLinks) {
			this._addDataAction(items, "about", "about", null, "global.action.title.about", "about", controller, overrides, visibleLinks);
		},
		_addDevButton: function(items, controller, overrides, visibleLinks) {
			this._addDataAction(items, "developmentPage", "developmentPage", null, "global.action.title.developmentPage", "development", controller, overrides, visibleLinks);
		},
		_addLogoutButton: function(items, controller, overrides, visibleLinks) {
			this._addDataAction(items, "logout", "logout", null, "global.action.title.logout", "logout", controller, overrides, visibleLinks);
		},
		_addDraftsListButton: function(items, numDrafts, controller, overrides, visibleLinks) {
			// Not sure if drafts will be done
			// this._addDataAction(items, "pageDrafts", "pageDrafts", numDrafts, "global.action.title.drafts", "drafts", controller, overrides);
		},
		_addAcceptButton: function(items, controller, overrides, visibleLinks) {
			this._addDataAction(items, "accept", "accept", null, "global.action.title.accept", "accept", controller, overrides, visibleLinks);
		},
		_addCancelButton: function(items, controller, overrides, visibleLinks) {
			this._addDataAction(items, "cancel", "cancel", null, "global.action.title.cancel", "cancel", controller, overrides, visibleLinks);
		},
		_addPageConfigButton: function(items, controller, overrides, visibleLinks) {
			this._addDataAction(items, "pageConfig", "pageConfig", null, "action.pageConfig", "pageConfig", controller, overrides, visibleLinks);
		},
		_addMyDashboardTemplatesButton: function(items, controller, overrides, visibleLinks) {
			this._addDataAction(items, "myDashboardTemplates", "myDashboardTemplates", null, "action.myDashboardTemplates", "myDashboardTemplates", controller, overrides, visibleLinks);
		},
		_addMyDashboardButton: function(items, controller, overrides, visibleLinks) {
			this._addDataAction(items, "myDashboard", "myDashboard", null, "action.myDashboard", "myDashboard", controller, overrides, visibleLinks);
		},
		// Title of action is "global.action.title." + action - optional params/icon
		_addGlobalAction: function(action, items, controller, overrides, icon, params, visibleLinks) {
			this._addDataAction(items, action, action, params, "global.action.title." + action, icon || action, controller, overrides, visibleLinks);
		},
		_addMultiSelectonButton: function(items, controller, overrides, visibleLinks) {
			this._addDataAction(items, "multiSelToggle", "multiSelToggle", null, "action.multiselection.toggle", "", controller, overrides, visibleLinks);
		},
		_addSortFilterButton: function(items, controller, overrides, visibleLinks) {
			this._addDataAction(items, "sortFilter", "sortFilter", null, "action.sortFilter", "sortFilter", controller, overrides, visibleLinks);
		},
		_addSearchButton: function(items, controller, overrides, visibleLinks) {
			this._addDataAction(items, "search", "search", null, "action.search", "search", controller, overrides, visibleLinks);
		},
		_addDraftsListLink: function(items, controller, slot) {
				// Add drafts
				this._addDataAction(items, "draftsList", "draftsList", null, "header.links.page.drafts", "drafts", controller, {
					slot: slot,
					disabled: true
				}, null);
			}
			//
			// END: Helpers
			//
	});

exports.createDataAction = function(name, action, params, title, icon, controller, overrides) {
	var item = $.extend(true, {
		"name": name,
		"controller-id": controller.id,
		"action": action,
		"params": params,
		"title": locale.textOpt(title) || title,
		"icon": icon
	}, overrides);
	return item;
}
