// syracuse-tablet/html/js/pages/pageDrafts.js
"use strict";

var globals = require('syracuse-tablet/html/js/helpers/globals');
var utils = require('syracuse-tablet/html/js/helpers/utils');
var Base = require('syracuse-tablet/html/js/pages/pageHtml').Page;
var locale = require('syracuse-tablet/html/js/helpers/locale');

var _pageTemplate = '\
		<div class="s-m-drafts-header">\
		</div>\
		{{#if isEmpty}}\
			<div>EMPTY LIST</div>\
		{{else}}\
			<div class="list-group s-m-drafts-list">\
			{{#each resources}}\
				<section id="{{id}}" class="list-group-item s-m-draft-item {{status}}">\
					<div style="display:table;table-layout:fixed;width:100%">\
						<div style="display:row">\
							<div style="display:cell">\
								<div><span>{{creation_date}}</span><span>{{title}}</span></div>\
								<div><span>{{endpoint}}</span><span>{{representation}}</span></div>\
								<div><span>{{reason}}</span><span>{{comment}}</span></div>\
							</div>\
							<div style="display:cell;width:100px">\
								<div style="width:50%">\
									<div data-action="delete" class="fa fa-times"></div>\
								</div>\
								<div style="width:50%">\
									<div data-action="edit" class="fa fa-pencil-square-o""></div>\
								</div>\
							</div>\
						<div>\
					<\div>\
				</section>\
			{{/each}}\
			</div>\
		{{\endif}}\
	';

var _Page = utils.defineClass(

	function DraftsPage($parent, state, options) {
		options.header = true;
		options.footer = false;
		Base.call(this, $parent, state, options);
		this._htmlTemplate = Handlebars.compile(_pageTemplate);
	}, Base, {
		scrollAllowed: function() {
			return true;
		},
		destroy: function() {
			Base.prototype.destroy.call(this);
		},
		appendHtml: function($$parent, loadOpts) {
			var self = this;
			self.$$elmt.addClass("drafts");
			globals.getStorage().draftOperation("draftReadList").then(function(resources) {
				$$parent.html(self._htmlTemplate, {
					resources: resources,
					isEmpty: resources == null || resources.length == 0
				});
			});
		},
		_actionEdit: function(e) {
			alert("_actionEdit");
		},
		_actDelete: function() {
			alert("_actDelete");
		}
	});

exports.Page = _Page;
