// syracuse-tablet/html/js/pages/pageHeader.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("page");
var templates = require('syracuse-tablet/html/js/helpers/templating');


/**
 * Page header
 *  Generate Html
 */
var _Klass = utils.defineClass(
	function($parent, page, prototype, $article) {
		var self = this;
		self.prototype = prototype;
		self.$article = $article;
		self._$parent = $parent;
		self.page = page;
	}, null, {

		destroy: function() {
			var self = this;
			self.prototype = null;
			self.$article = null;
			self.page = null;
			if (!self.$elmt) return;
			utils.unbindObj(self.$elmt);
			self.$elmt = null;
			utils.unbindObj(self);
		},

		loadStructure: function() {},

		buildHtml: function() {
			var self = this;
			if (self.page.vignette) {
				return null;
			} else {
				return templates.execSync("contructionHeadTmpl", {
					title: "Sage X3 Mobile",
					pageId: self.page.id,
					logout: "logout"
				});
			}
		}
	});

exports.Klass = _Klass;
