// syracuse-tablet/html/js/pages/pageDashboard.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var Base = require('syracuse-tablet/html/js/pages/page').Page;
var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("page");
var uiRect = require('syracuse-tablet/html/js/ui/rect');

var _Page = utils.defineClass(

	function DashboardPage($parent, state, prototype, article) {
		var self = this;
		Base.call(self, $parent, state, prototype, article);
		self.vignettes = [];
	}, Base, {

		destroy: function() {
			var self = this;
			Base.prototype.destroy.call(self);
			self.vignettes = null;
		},

		loadData: function(options) {
			return null;
		},

		_beforeAddContent: function() {
			var self = this;
			self.addTitle(self.prototype.data("$title"), self.prototype.data("$description"));
		},

		addControl: function(c, isPageControl) {
			var self = this;
			Base.prototype.addControl.call(self, c, isPageControl);
			if (c && c.isVignette()) {
				self.vignettes.push(c);
			}
		},


		/**
		 * Scan vignettes to find control per id
		 * 	Allows to call ctrl actions in vignettes
		 * ctrlId is a unicid for all controls of all pages
		 */
		findVignetteControl: function(ctrlId) {
			var self = this;
			if (!self.vignettes) return null;
			var c, v;
			for (var i = 0; i < self.vignettes.length; i++) {
				v = self.vignettes[i];
				c = v.getPageControl(ctrlId);
				if (c) {
					return {
						control: c,
						vignette: v
					};
				}
			}
			return null;
		},
		/**
		 * Scan vignettes to find page per id
		 */
		findVignettePage: function(pageId) {
			var self = this;
			if (!self.vignettes) return null;
			for (var i = 0, v; i < self.vignettes.length; i++) {
				v = self.vignettes[i];
				if (v.getPage() && v.getPage().id === pageId) {
					return {
						page: v.getPage(),
						vignette: v
					};
				}
			}
			return null;
		},
		afterRender: function(options) {
			var self = this;
			var deferred = $.Deferred();
			var rejectError = function(e) {
				log && log("Error loading vignettes", e);
				deferred.reject(e);
			};
			try {
				/* Standard process - bind actions - no deferred returned*/
				Base.prototype.afterRender.call(self, options);
				/* Load vignettes*/
				var promises = [];
				self.vignettes.forEach(function(v) {
					var options = {
						sdataParameters: self.state.options && self.state.options.sdataParameters || {}
					};
					v.load(null, options);
				});
				deferred.resolve();
			} catch (e) {
				rejectError(e);
			} finally {
				return deferred.promise();
			}
		},
		/**
		 * Context to store when dashboard is destroyed
		 * We save the context for each vignette (scroll, selection....)
		 * Context of the vignette will be retrieved with the index which is the identifier
		 * Vignette id is recalculated at creation and changes each time the vignette is loaded
		 */
		savedCtxCreate: function() {
			var self = this;
			var ctx = Base.prototype.savedCtxCreate.call(self);
			ctx.vignettes = [];
			var p;
			self.vignettes && self.vignettes.forEach(function(v) {
				p = v.getPage();
				// We store a context even if p == null - Vignette can be destroyed before the end of load process and we must preserve the index in vignette array
				// Vignette is identified by its index
				ctx.vignettes.push(p ? p.savedCtxCreate() : null);
			});
			if (this.rootLayout && this.rootLayout.getGestureMgr()) {
				// Store the scroll of hub layout
				ctx.hubScroll = this.rootLayout.getGestureMgr().savedCtxCreate();
			}
			return ctx;
		},
		/**
		 * Restore the context for the dashboard (not for the vignette)
		 */
		savedCtxRestore: function(ctx) {
			// Std restore for all page
			Base.prototype.savedCtxRestore.call(this, ctx);
			// store vignettes context
			this._savedVignetteContext = ctx.vignettes;
			if (ctx.hubScroll && this.rootLayout && this.rootLayout.getGestureMgr()) {
				// Restore the scroll of hub layout
				this.rootLayout.getGestureMgr().savedCtxRestore(ctx.hubScroll);
			}
		},
		/**
		 * Returns the context of a vignette v
		 * Called by the vignette
		 */
		savedCtxVignette: function(v) {
			var self = this;
			if (v == null || this._savedVignetteContext == null) return null;
			var idx = self.vignettes.indexOf(v);
			if (idx == -1 || idx > this._savedVignetteContext.length - 1) return null;
			return {
				current: this._savedVignetteContext[idx]
			};
		},
		/**
		 * Scrolling is handled by layout hub
		 */
		scrollAllowed: function() {
			return false;
		}
	});

exports.Page = _Page;
