// syracuse-tablet/html/js/pages/page.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var globals = require('syracuse-tablet/html/js/helpers/globals');
var Base = require('syracuse-tablet/html/js/helpers/controller').Klass;;
var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("page");
var uiutils = require('syracuse-tablet/html/js/ui/uiUtils');
var uiRect = require('syracuse-tablet/html/js/ui/rect');
var ctrlFactory = require('syracuse-tablet/html/js/controls/ctrlFactory');
var HammerScroller = require('syracuse-tablet/html/js/helpers/hammerScroller').Klass;
var settings = require('syracuse-tablet/html/js/application/settings');
var notifications = require('syracuse-tablet/html/js/helpers/notifications');

/**
 * s-m-side-panel-dismiss: div that traps the events outside the panel when a panel if open
 */
var _templates = {
	fullpage: '\
		<section class="s-m-page s-m-full {{type}} {{subtype}}" id="{{pageid}}" {{#if parentId}}data-parent-id="{{parentId}}"{{/if}} class="ctrl-event-click" style=""> \
			<header></header> \
			<section class="s-m-main-content"></section> \
			<footer></footer> \
		</section> \
	',
	vignette: '<section class= "s-m-page s-m-nested {{type}} {{subtype}}" id="{{pageid}}"></section>',
	panels: '\
		<div class="s-m-side-panel-dismiss" data-action="toggleSidePanel" data-params="close"></div> \
		<aside class="s-m-panel s-m-left"></aside> \
		<aside class="s-m-panel s-m-right"></aside>'
};
var _getHtml = function(name, ctx) {
	var key = name + "compiled";
	var tmpl = _templates[key];
	if (!tmpl) {
		tmpl = _templates[key] = Handlebars.compile(_templates[name]);
	}
	return tmpl(ctx || {});
};
/*********** PANELS MANAGER ***********
 * !! Created/Destroyed only on main page (not on vignettes)
 * page		Main/Current page
 *************************************/
var PanelManager = function(page) {
	this.page = page;
	if (!page.isVignette) {
		// init $$elmts for panels
		page.$$elmt.append($(_getHtml("panels")));
		page.$$dismissPanel = page.$$elmt.children(".s-m-side-panel-dismiss");
	}
	this.destroy = function() {
		this.page = null;
	};
	/**
	 * Panels are attached to the main page
	 */
	this.mainPage = function() {
		return this.page.isVignette ? globals.getCurrentPage() : this.page;
	};
	/**
	 * Adds a panel to left or right sections - sidePanel class
	 * 		If not exists create it
	 * where	"left" or "right"
	 * panel	panel object
	 */
	this.panelAttach = function(where, panel, action) {
		var controller = this.mainPage();
		if (!controller.$$dismissPanel || !panel || (where != "right" && where != "left")) return;
		if (controller._pageControls.get(where) && controller._pageControls.get(where) != panel) {
			this._panelDetach(where);
		}
		if (!panel.isAttached(where)) {
			var $$parent = controller.$$elmt.children("aside.s-m-panel.s-m-" + where);
			if ($$parent.length != 1) return;
			// Add the control to the list
			controller.controlsMap[panel.id] = panel;
			controller._pageControls.add(where, panel).attach($$parent, where);
		}
		return this.panelToggle(where, panel, action);
	};
	/**
	 * Toggle panel
	 */
	this.panelToggle = function(where, panel, action) {
		var controller = this.mainPage();
		if (!panel) panel = controller._pageControls.get(where);
		if (!panel || !panel.$$elmt) return null;
		// close -> Force hide - resize -> stay visible - toggle -> toggle
		panel.$$elmt.toggle(action === "close" ? false : undefined);
		var visible = panel.$$elmt.is(':visible') === true;
		var newWidth = visible ? panel.$$elmt.width() : 0;
		if (panel.closeBox != true) {
			controller.$$dismissPanel[visible ? "addClass" : "removeClass"](where);
			controller.$$dismissPanel.css(where, newWidth);
		}
		if (panel.overlap != true) {
			var self = this;
			var shiftElmts = [controller.$$contentElmt];
			["header", "footer"].forEach(function(id) {
				var c = controller._pageControls.get(id);
				if (c) {
					shiftElmts.push(c.$$elmt.parent());
				}
			});
			shiftElmts.forEach(function($$elmt) {
				$$elmt.css("left", (visible ? (where === "right" ? -1 : 1) * newWidth : 0) + "px");
			});
		}
		panel.notifyToggled(where, action, visible);
		return panel;
	};
	/**
	 * Removes the panel but not destroy it
	 * where	"left" or "right"
	 */
	this._panelDetach = function(where) {
		var controller = this.mainPage();
		if (!controller.$$dismissPanel) return;
		var p = controller._pageControls.remove(where);
		if (!p) return;
		this.panelToggle(where, p, "close");
		// Not undefined because there are loops on controlsMap
		delete controller.controlsMap[p.id];
		return p.detach();
	};
	/**
	 * Returns a panel - Just to have ctrlFactory declared here
	 * id	pageactionpanelregular	 	for action panel
	 * 		userpanelsmartphone			for left panel
	 */
	this.panelCreate = function(panelId, options) {
		// controller of the control is the current page because we could need to acced to this page from the panel (Ex for a panel attached to a vignette)
		return ctrlFactory.createPanel(this.page, panelId, options);
	};
	this.panelResize = function() {
		var controller = this.mainPage();
		["right", "left"].forEach(function(w) {
			var p = controller._pageControls.get(w);
			if (p && p.onResize) p.onResize();
		});
	};
};
/*********** PAGE CONTROLS MANAGER ***********
 * Page controls are not attached to rootLayout
 * header/footer
 ************************************************/
var PageControls = function(page) {
	this.page = page;
	this.map = {};
	this.statics = {};
	this.add = function(id, ctrl) {
		if (this.map[id]) {
			alert("Control " + id + " already exists");
		}
		this.map[id] = ctrl;
		if (id === "header" || id === "footer") {
			this.statics[id] = ctrl;
		}
		return ctrl;
	};
	this.get = function(id) {
		return this.map[id];
	};
	this.remove = function(id) {
		var ctrl = this.map[id];
		this.map[id] = undefined;
		return ctrl;
	};
	this.destroy = function() {
		this.page = null;
		if (!this.map) return;
		for (var id in this.map) {
			this.map[id] && this.map[id].destroy();
		}
		this.map = null;
		this.statics = null;
	};
	/**
	 * Only on static controls header/footer
	 */
	this.buildHtml = function(dao, buildOptions) {
		for (var id in this.statics) {
			this.statics[id].buildHtml(this.page.$$elmt.children(id), dao, buildOptions);
		}
	};
	this.afterRender = function(buildOpts) {
		for (var id in this.statics) {
			if (this.statics[id].afterRender) this.statics[id].afterRender(buildOpts);
		}
	};
	/**
	 * All controls header/footer/leftPanel/rightPanel
	 */
	this.forAllCtrls = function(cb) {
		for (var id in this.map) {
			cb(id, this.map[id]);
		}
	};
	this.addToControlMap = function() {
		for (var id in this.map) {
			this.page.addControl(this.map[id], true);
		}
	};
	this.closeSidePanel = function() {
		// TODO
	};
};
/****************** PAGE CLASS *********************
 * Base class for all pages
 *  $$parent: Parent element
 * 	state: Page information - JSON structure stored in history
 *  prototype: Prototype object (class prototype)
 *  article: JSON article
 *  !! Temporarily prototype and article can be null for html pages
 ***************************************************/
var _defOpts = {
	header: true,
	footer: true
};
var _Page = utils.defineClass(
	function Page($$parent, state, prototype, article, options) {
		Base.call(this, state.uuid, prototype, article, $.extend({}, _defOpts, options));
		this._pageControls = null;
		this._panelsManager = null;
		this._$$parent = $$parent;
		this.state = state;
		this.state.options = state.options || {};
		this._initMaps();
		this.isVignette = state.options.vignette === true;
		this.isChild = state.options.parentId != null;
		if (this.isChild) {
			// Children pages are cached - no destry if we open a child page from this page
			this.state.cached = true;
		}
		this._initRootElmt($$parent);
		// true if page is created - set by page loader when page is activated
		this.isNew = false;
		// Last selected row - Used to save/restore context of last selected row
		this.lastSelectedRow = null;
		notifications.subscribe(this, ["sm.select.row"], 1);
	}, Base, {
		_initMaps: function() {
			// Contains all the controls of the page including page controls
			this.controlsMap = {};
			// Contains controls that need to be called after Render - outside pageControls
			this.ctrlAfterRenderCalls = [];
			// Contains controls that have a context to save in order to retrieve it when page will be load (eg: back navigation)
			this.ctrlSaveCtxCalls = [];
		},
		/**
		 * Create and return root DOM element of every page
		 * 		DOM element is hidden and contains only the structure
		 * 		Page object is stored in DOM element under data("smcontroller")
		 */
		_initRootElmt: function($$parent) {
			var self = this;
			var type = self.state.type;
			var root, ctx = {
					pageid: self.id,
					type: type, // regular/html/dashboard
					subtype: self.state.subtype, // regular - > facet - dashboard -> std/welcome/home - html -> login/about...
					parentId: self.getParentId()
				};
			if (self.isVignette) {
				// only content in a vignette
				self.$$contentElmt = self.$$elmt = $(_getHtml("vignette", ctx));
			} else {
				self.$$elmt = $(_getHtml("fullpage", ctx));
				self.$$contentElmt = self.$$elmt.children("section.s-m-main-content");
				// Page controls only for main page
				self._pageControls = new PageControls(self);
				if (self.options.header === true) {
					// It's better to fix the top before mainResize even if it can change
					// When we restore context (scroll...) on back we need that because restore context is done before mainREsize
					self.$$contentElmt.css({
						top: globals.getSiteLayout().getHeight("header")
					});
				}
			}
			// Panel manager should be attached to the main page - For vignettes we should take toe one of the main page
			//  -> Temporarlilly we create a Panel manager for vignettes because main page (globals.currentPage) is not available when we create the controls of the vignette
			//  -> Ex when we create an array sortFilter panel in a vignette
			//  -> TODO - Review the loading process of vignettes to pass the main page (pageToOpen) as a parameter or option 
			//  -> After that we will be able to set self._panelsManager = mainPage._panelsManager for vignettes
			self._panelsManager = new PanelManager(self);
			if (self.scrollAllowed()) {
				// Scroll in page is allowed for $details and $edit
				// For dashboard this is the layout which scrolls (not the page)
				// For query this si the array  which scrolls (not the page)
				// Only if scroll allowed - Ex: In wignette with charts we musn't have the scroll wrapper becaus eit breaks size calculation 
				self.$$scrollWrapper = $('<div class="s-m-scroll-wrapper"/>').appendTo(self.$$contentElmt);
				self.$$scrollWrapper.css({
					height: "auto"
				});
				self.$$ctrlsRootElmt = self.$$scrollWrapper;
				// Create scroll manager here because it must be created when we restore the context on back
				self._gestureMgr = new HammerScroller(self.$$scrollWrapper, {
					// isPageScroller: true for full page
					// isPageScroller: false for vignette - The view rectangle scolls with the content of the page
					isPageScroller: !self.isVignette
				});
			} else {
				self.$$ctrlsRootElmt = self.$$contentElmt;
			}
			self.$$elmt.hide();
			self.$$elmt.appendTo($$parent);
			// attaches page object to DOM element
			self.$$elmt.smPageController(self);
		},
		isDashboard: function() {
			return this.getType() === "dashboard";
		},
		destroy: function() {
			Base.prototype.destroy.call(this);
			this._$$parent = null;
			if (this.$$elmt) {
				utils.unbindObj(this.$$elmt);
				this.$$elmt = null;
			}
			if (this._pageControls) {
				this._pageControls.destroy();
				this._pageControls = null;
			}
			if (this._panelsManager) {
				this._panelsManager.destroy();
				this._panelsManager = null;
			}
			if (this._waiting) {
				uiutils.waitWheelDestroy(this._waiting);
				this._waiting = null;
			}
			this._initMaps();
			if (this._gestureMgr) {
				this._gestureMgr.destroy();
				this._gestureMgr = null;
			}
		},
		/**
		 * Called be pageLoader - TODO TEST
		 */
		destroySiblings: function() {
			if (this.$$elmt) {
				log && log("Page destroySiblings", "destroySiblings");
				var next = this.$$elmt.siblings(function(idx) {
					var ctrl = $(this).smPageController();
					if (ctrl) ctrl.destroy();
				});
			}
		},
		/**
		 * dashboard/regular/html
		 */
		getType: function() {
			return this.state.type;
		},
		/**
		 * Allows prop = "options.parentId"
		 */
		getProp: function(prop) {
			return utils.getPropByPath(this.state, prop);
		},
		/**
		 * Set by vignette at page loading
		 */
		openLinkInVignette: function() {
			return this.state.options.openLinkInVignette === true;
		},
		/**
		 * Load the page
		 * 		The result is that the page is created and attached to DOM document and hidden
		 * 		the contoller of the page (applictaion/vignette) has to call activate method to display the page (see pageLoader module)
		 * updateLayout:
		 * 		true to not load the data and update only main layout
		 */
		load: function(updateLayout) {
			var self = this;
			var updateLayout = updateLayout === true;
			// We store buildOpts
			var deferred = $.Deferred();
			var timeLine = utils.getTimeLine("page", self.state.name);
			timeLine && timeLine.start();
			try {
				self.buildHtmlOption()
					.then(function(opts) {
						self.buildOpts = opts;
					})
					.then(function() {
						return self.loadStructure(updateLayout);
					})
					.then(function() {
						if (updateLayout) {
							return $.smResolve();
						}
						timeLine && timeLine.step("loadStructure");
						return self.loadData();
					})
					.then(function() {
						timeLine && timeLine.step("loadData");
						return self.render(updateLayout, self.buildOpts);
					})
					.then(function() {
						timeLine && timeLine.step("render");
						return self.afterRender(updateLayout, self.buildOpts);
					})
					.then(function() {
						deferred.resolve();
						timeLine && timeLine.stop();
					}).fail(function(e) {
						deferred.reject(e);
					});
			} catch (e) {
				deferred.reject(e);
			} finally {
				return deferred.promise();
			}
		},
		/**
		 * Create layout/control tree structure according to article
		 */
		loadStructure: function(updateLayout) {
			var deferred = $.Deferred();
			try {
				if (!this.article || !this.prototype) throw new Error("loadStructure - NULL article or prototype");
				if (this._pageControls && updateLayout !== true) {
					if (this.options.header === true) {
						this._pageControls.add("header", ctrlFactory.createPageControl("pageheader", this));
					};
					if (this.options.footer === true) {
						this._pageControls.add("footer", ctrlFactory.createPageControl("pagefooter", this));
					}
				}
				// Creates control / fields tree(from article)
				this.$$elmt.attr("data-layout-path", (this.state.options.layoutPath || 0) + "#page");
				this.buildControlsTree(this.state.options.layoutPath);
				deferred.resolve();
			} catch (e) {
				deferred.reject(e);
			} finally {
				return deferred.promise();
			}
		},
		/**
		 * Load page data
		 */
		loadData: function() {
			var deferred = $.Deferred();
			var self = this;
			try {
				self.setDao(null);
				deferred.resolve();
			} catch (e) {
				deferred.reject(e);
			} finally {
				return deferred.promise();
			}
		},
		/**
		 * Rendering - Build html and attachs dom elements to document
		 */
		render: function(updateLayout, buildOpts) {
			try {
				if (!this.rootLayout) {
					throw new Error("buildHtml - NULL rootLayout");
				}
				// Renders header/footer
				if (this._pageControls && updateLayout !== true) {
					this._pageControls.buildHtml(this.dao, buildOpts);
				}
				if (this.$$ctrlsRootElmt) {
					this.beforeAddContent();
					// Renders tree controls (article)
					Base.prototype.render.call(this, this.$$ctrlsRootElmt, buildOpts);
				}
			} catch (e) {
				return $.smReject(e);
			}
			return $.smResolve();
		},
		/**
		 * return build options for controls
		 * Set options for all controls
		 * Header -> options for header control
		 * Can be overridden
		 */
		buildHtmlOption: function() {
			var self = this;
			var allowPin = true;
			var $typ = self.prototype && self.prototype.data("$typ");
			if ($typ == "rowdetail") {
				allowPin = false;
			}
			return settings.isPinnedPage(self.getPinPageId())
				.then(function(isPinned) {
					return {
						makeFullHeight: this.isVignette,
						toolBars: {
							showAuthoring: !this.isChild && this.state.disableAuthoring !== true,
							showLinks: !this.isChild,
							showHome: !this.isChild && !this.state.isWelcomePage,
							statusPinPage: allowPin ? (isPinned ? 2 : 1) : 0
						}
					};
				});
		},
		/**
		 * Called from authoring
		 *	TODO Review FDB - use pageNotification or reload the page
		 */
		updateLayout: function(article) {
			var self = this;
			// Set new article which comes from authoring panel
			self.article = article;
			// Init controsl maps
			self._initMaps();
			// destroy all controls and layouts
			self.rootLayout.destroy();
			// Clear dom
			self.$$ctrlsRootElmt.empty();
			// Reload with updateLayout option
			// 	-> don't load data
			//	-> don't create page controls
			return self.load(true).then(function() {
				// Add page controls to maps
				self._pageControls.addToControlMap();
				// Recalculate layouts
				uiutils.triggerResizeInternal();
			});
		},
		/**
		 * Resize of application main page
		 * info give the size in pixels of mainContent/header/footer - see siteLayout
		 * We need it to pass to rootLayout without having to set the size od $$contentElmt
		 * $$contentElmt must have css.height set to auto for scrolling
		 */
		onMainPageResize: function(info, orientation, deviceClass) {
			var self = this;
			if (!this.isVignette) {
				// Set device type for css
				if (this._deviceType) this.$$elmt.removeClass(this._deviceType);
				this._deviceType = globals.getSiteLayout().getDeviceType();
				this.$$elmt.addClass(this._deviceType);
				/**
				 * or main page we calculate the H/W in pixels - BAs of calculation comes from application/SiteLayout
				 */
				var top = 0,
					left = 0;
				var contentHeight = info.application.height;
				var contentWidth = info.application.width;
				if (self._pageControls) {
					var ctx = {
						application: info.application,
						contentWidth: contentWidth,
						contentHeight: contentHeight
					};
					self._pageControls.forAllCtrls(function(id, ctrl) {
						if (ctrl.onMainPageResize) {
							ctrl.onMainPageResize(ctx, orientation, deviceClass);
							if (ctrl.isVisible() && (id === "header" || id === "footer")) {
								contentHeight -= ctrl.getHeight();
								if (id === "header") top += ctrl.getHeight();
							}
						}
					});
				}
				self.$$contentElmt.css({
					top: top + "px",
					left: left + "px",
					width: contentWidth + "px",
					height: contentHeight + "px"
				});
			} else {
				/**
				 * For vignette we don't calculate the H/W in pixels.
				 * To do the calculatation we need to propagate the calculation from dashboard tile -> Control (not done and needed)
				 * H/W calculation is based on css style height auto for the wrapper et 100% for $$contentElmt
				 */
			}
			/**
			 * We need to create the scroller for main page and page embeded in vignette
			 */
			// Save content rectangle for scrolling and hub layout - "computed" to ignore padding
			self._mainContentRect = uiRect.elmtRect(self.$$contentElmt, "computed");
			// We need to set the right coordinates for intersectRect with scroller
			self._mainContentRect.moveLeft(parseInt(self.$$contentElmt.css("padding-left"), 10));
			this._initScroll();
			self.onResize();
		},
		/**
		 * returns main content rect for hub layout
		 */
		getMainContentRect: function() {
			return this._mainContentRect;
		},
		/**
		 * Resize handler called by window.resize
		 */
		onResize: function() {
			// Call resize handler in tree hierachy
			if (this.rootLayout && this.rootLayout.onResize) {
				this.rootLayout.onResize();
			}
			if (this._panelsManager) {
				this._panelsManager.panelResize();
			}
		},
		/**
		 * Specific procces after rendering
		 */
		afterRender: function(updateLayout, buildOpts) {
			// afterRender for controls - outside pageControls
			this.ctrlAfterRenderCalls.forEach(function(c) {
				c.afterRender(buildOpts);
			});
			if (this._pageControls && updateLayout !== true) {
				// for _pageControls afterRender is called only at the creation of the page
				this._pageControls.afterRender(buildOpts);
			}
		},
		beforeAddContent: function() {},
		addTitle: function(title, description) {
			var html = "";
			if (title) html = '<header class="s-m-page-title">' + title + '</header>';
			if (description) html += '<p>' + description + '<p>';
			if (html.length > 0) $(html).prependTo(this.$$ctrlsRootElmt);
		},
		/**
		 *  called each tile a control/layout is created
		 *  isPageControl: 	true is it's a page control - Panel, Header, Footer...
		 **/
		addControl: function(c, isPageControl) {
			Base.prototype.addControl.call(this, c);
			// For controls others that page Controls
			if (isPageControl !== true) {
				// stores controls that need specific processes
				// !!! It's the simplest way to identify controls by testing the presence of the method when we add them
				// Another way woud be that it's the control that asks the page to be notified
				// After render
				if (c.afterRender) this.ctrlAfterRenderCalls.push(c);
				// Save context
				if (c.savedCtxCreate && c.savedCtxRestore) this.ctrlSaveCtxCalls.push(c);
			} // PageControls are managed separately because of updateMayout 
		},
		/**
		 * Display the page and hides currentPage
		 * 		Manage transitions (effect) according  to self.getProp("transition")
		 *		Do not call show directly
		 * 		cb is called after oage hs been displayed (we don't use deferred)
		 * pageToClose is the page ths will be replaced by this
		 * 		For a child page currentPage is the parent page
		 *			Used by rowdetail page to read the DAO
		 * savedContext
		 * 		Context is propvided by pageToClose (or dashboard for vignettes)
		 * 		It contains a status to restore after it has been displayed (scroll...)
		 */
		activate: function(pageToClose, savedContext, cb) {
			var self = this;
			self._show(pageToClose, function() {
				if (savedContext) {
					// Set saved context use for back to 'parent' page
					self.savedContext = savedContext.parent;
					// We restore the current contex (ex scroll)
					self.savedCtxRestore(savedContext.current);
				} else if (self.lastSelectedRow) {
					// We allways restore last selected row because it is stored in page when we clik the link that opens the 'child' page
					self._hilightLastSelectedRow(self.lastSelectedRow);
				}
				if (cb) cb();
			});

		},
		/**
		 * Hide the page
		 * 	Do not call hide or show directly
		 * A "cached" page is deactivated instead of being destroyed on navigation event
		 */
		deactivate: function(cb) {
			this._hide(cb);
		},
		/**
		 * Save previous page context - Called by page loader before deactivating the page
		 * We propagate the context of previous pages into the page to be able to retrieve it on back navigation
		 */
		contextSave: function(pageToClose) {
			var self = this;
			if (!pageToClose) return;
			self.savedContext = {
				// Parent context is saved
				parent: pageToClose.savedContext,
				// Context to restore when this pageToClose will be reopen on back navigation
				current: pageToClose.savedCtxCreate()
			};
		},
		/**
		 * returns the context to save
		 * 		default:	page scroll coordinates
		 * Can be overridden (ex dashboard to manage contexts of vignettes)
		 */
		savedCtxCreate: function() {
			var self = this;
			var ctx = {
				pageName: this.state.name,
				lastSelectedRow: self.lastSelectedRow
			};
			if (self._gestureMgr) {
				// Save page scrolling
				ctx.scroll = self._gestureMgr.savedCtxCreate();
			}
			if (self.ctrlSaveCtxCalls.length > 0) {
				ctx.ctxCtrls = {};
				// Save controls contexts - All controls who implement savedCtxCreate are added to ctrlSaveCtxMap
				// We take $bind as control ID because id is changed each time we create a control
				self.ctrlSaveCtxCalls.forEach(function(c) {
					var ctrlCtx = c.savedCtxCreate();
					if (ctrlCtx) ctx.ctxCtrls[c.$bind] = ctrlCtx;
				});
			}
			return ctx;
		},
		/**
		 * Restore page context - Called after activation (show)
		 * 		default:	page scroll coordinates
		 */
		savedCtxRestore: function(ctx) {
			var self = this;
			if (!ctx) return;
			if (ctx.pageName != this.state.name) {
				// The page is not able to restore child pages inside a vignette - TODO if needed
				// If we go to detail page in ful page from a dashboard and go back a vignette that contains a detail page will not restore the context if current page was a child page
				// -> Only main parent page is restored
				return;
			}
			if (self._gestureMgr && ctx.scroll) {
				// Restore page scrolling
				self._gestureMgr.savedCtxRestore(ctx.scroll, true);
			}
			if (ctx.ctxCtrls) {
				// Restore controls context
				// We take $bind as control ID because id is changed each time we create a control
				for (var bind in ctx.ctxCtrls) {
					var c = self.getControlByBind(bind);
					var ctrlCtx = ctx.ctxCtrls[bind];
					if (c && c.savedCtxRestore && ctrlCtx) {
						if (ctx.lastSelectedRow && ctx.lastSelectedRow.array$Bind === bind) {
							// We pass the id to the right arrays - Row selection will be done later after the display of th array
							ctrlCtx.selectedRowId = ctx.lastSelectedRow.rowId;
						}
						c.savedCtxRestore(ctrlCtx);
					}
				}
			}
		},
		_hilightLastSelectedRow: function(lastSelectedRow, old) {
			var self = this;
			var toggle = function(rowInfo, show) {
				// We take array$Bind as control ID because id is changed each time we create a control
				var c = self.getControlByBind(rowInfo.array$Bind);
				if (c && c.selectLastSelectedRow) {
					c.selectLastSelectedRow(rowInfo.rowId, show);
				}
			};
			if (old) toggle(old, false);
			if (lastSelectedRow) {
				toggle(lastSelectedRow, true);
				return true;
			}
			return false;
		},
		_show: function(currentPage, cb) {
			var self = this;
			var transition = self.getProp("transition");
			if (currentPage) {
				currentPage._hide(function() {
					self.$$elmt.show(transition || 0, function() {
						if (cb) cb();
					});
				});
				return;
			}
			self.$$elmt.show(transition == null ? 0 : transition, cb);
		},
		_hide: function(cb) {
			if (!this.$$elmt) {
				if (cb) cb();
				// page has been destroyed
				return;
			}
			// Close panels
			if (this._pageControls) {
				this._pageControls.closeSidePanel("all");
			}
			var transition = this.getProp("transition");
			this.$$elmt.hide(transition == null ? 0 : transition, cb);
		},
		isActive: function() {
			return this.$$elmt && this.$$elmt.is(":visible");
		},
		/**
		 * True is page is a lookup
		 */
		isLookup: function() {
			return this.prototype && this.prototype.isLookup();
		},
		/**
		 * Return the info to notify parent when the child page is closed
		 * It depends on the kind of page
		 * Ex: lookup notifies parent control with selected value
		 */
		getParentNotif: function() {
			throw new Error("not implemented");
		},
		/**
		 * Return the id of parent page if any
		 */
		getParentId: function() {
			return this.isChild ? this.state.options.parentId : null;
		},

		/**
		 * Returns attribute for the link that opens this page (used to open this page from a vignette to full page)
		 * Overriden by regular page to put s-data-url
		 */
		getOpenLinkAttrs: function() {
			return {
				"data-nav": this.state.name
			};
		},

		/***
		 * Test action
		 */
		_actTest: function(params) {
			var e = $('<h5 style="color:red">Action test succeeded</h5>').prependTo(this.$$elmt);
			window.setTimeout(function() {
				e.remove();
			}, 2000);
		},
		/**
		 * Click on home in footer/header
		 * Can be overrridden - rowDetail
		 */
		_actHome: function() {
			globals.getApplication().gotoWelcomeApplication();
		},
		/**
		 * Wait wheel management
		 * We need one plugin per page
		 */
		waitWheelStart: function(controlId) {
			var self = this;
			if (controlId) {
				var c = self.controlsMap[controlId];
				if (c && c.waitStart) c.waitStart();
				return;
			}
			if (!self._waiting) {
				self._waiting = uiutils.waitWheelCreate(self.$$elmt);
			}
			uiutils.waitWheelStart(self._waiting);
		},
		waitWheelStop: function(controlId) {
			var self = this;
			if (controlId) {
				var c = self.controlsMap[controlId];
				if (c && c.waitStop) c.waitStop();
				return;
			}
			if (self._waiting) {
				uiutils.waitWheelStop(self._waiting);
			}
		},
		/**
		 * Called synchronously by an array when a row has been selected
		 * 	arrayId		Id of array
		 * 	rowId		Id of row
		 * 	rowdata		data of the row used by loojup page
		 * 	updateUI	True to update UI
		 * 				When the page is not destroyed by the link we clicked we have to remove old selection
		 * 				Otherwise the selction the line will be hightlighted after creation of the page
		 */
		notifSelectRow: function(arrayId, rowId, rowdata, updateUI) {
			var old = this.lastSelectedRow;
			this.lastSelectedRow = {
				array$Bind: this.getControl(arrayId).$bind,
				rowId: rowId
			};
			if (updateUI) {
				this._hilightLastSelectedRow(this.lastSelectedRow, old);
			}
		},
		/**
		 * USed by child pages
		 * TODO - Improve by using orr own history stack to avoid forward
		 */
		goBack: function() {
			window.history.back();
		},
		/**
		 * Update the scroll for a vignette
		 *
		 */
		initVignetteScroll: function(back) {
			if (!this._gestureMgr) return;
			var viewRect = this.getParentVignette().getScrollViewRect();
			//if (back === true || (viewRect.height < this.$$scrollWrapper.height())) {
			// If back into a vignette we keep old scrolling - To improve 
			var keepOldScrolling = back === true;
			this._gestureMgr.init(viewRect, "v", 0, keepOldScrolling);
			//}
		},
		/**
		 * Enable/Disable scrolling
		 */
		_initScroll: function(back) {
			if (!this._gestureMgr) return;
			if (this.scrollAllowed()) {
				var viewRect = this.isVignette ? this.getParentVignette().getScrollViewRect() : this._mainContentRect;
				// Force $$scrollWrapper to full heigth of childrens
				// We scroll if childen height is greater than childrenRect height
				if (back === true || (viewRect.height < this.$$scrollWrapper.height())) {
					var valMax = 0;
					// If not new -> back on a cahced page - We keep scrolling
					var keepOldScrolling = this.isNew !== true;
					//viewRect has fixed coordinate to check if the 'finger' is inside on mousemouve event
					// A page always scolls verticaly - In dashboard this is the layout that scrolls horizontally
					this._gestureMgr.init(viewRect, "v", valMax, keepOldScrolling);
					var autoScroll = this.state.options["auto-scroll"];
					if (autoScroll != null) {
						// When we go to $edit form $details we keep the scrolling in order to stay at the same location (see actionManager)
						this.autoScroll(autoScroll);
						// Only when we loadd the page
						this.state.options["auto-scroll"] = undefined;
					}
				}
			} else {
				// Disable _gestureMgr if any
				this._gestureMgr.reset();
			}
		},
		/**
		 * Scroll the page
		 */
		autoScroll: function(value) {
			// TODO - scroll the rootLayout for the dashboard
			if (this._gestureMgr) this._gestureMgr.autoScroll(value);
		},
		/**
		 * True if scroll enabled in page
		 * By default we can scroll a page in full display
		 * In vignettes only details/edit facet are allowed to scroll - See regular page
		 * In Dashboard hub this is the layout that scrolls - Page doesn't scroll
		 */
		scrollAllowed: function() {
			if (this.isVignette) return false;
			if (this.isDashboard()) return this.article && this.article.$layoutType !== "hub";
			return true;
		},
		/**
		 * Returns the offset of the scrollRect top/left in fixed position (document)
		 */
		getScrollOffset: function() {
			// Application content can be scrcolled (native in authoring)
			var offset = globals.getApplication().getScrollOffset();
			if (this.isDashboard()) {
				// For readability t's better to put this code here than in dasborad class
				// return hub dashboard H or V shifting
				var g = this.rootLayout ? this.rootLayout.getGestureMgr() : null;
				if (!g) return offset;
				var o = g.getOffset();
				offset.top += o.top;
				offset.left += o.left;
				return o;
			} else {
				// No offset
				return offset;
			}
		},
		/**
		 * returns the value of the page's scroller
		 * 0 by default even if no scroller		 *
		 */
		getScrollValue: function() {
			return this._gestureMgr ? this._gestureMgr.getScrollValue() : 0;
		},
		setParentVignette: function(v) {
			this._parentVignette = v;
		},
		getParentVignette: function() {
			return this._parentVignette;
		},
		getMainPageGestureMgr: function() {
			var p = globals.getCurrentPage();
			if (p.isDashboard()) {
				// It's the hub layout that scrolls - not  the page
				// For readability t's better to put this code here than in dasborad class
				return p.rootLayout ? p.rootLayout.getGestureMgr() : null;
			} else {
				return p._gestureMgr;
			}
		},
		/**
		 * Show element in page - scroll if needed
		 * c is a JQuery elmt or an object with $$elmt property
		 */
		makeVisible: function(c) {
			if (!c || !this._gestureMgr || !c.$$elmt) return;
			this._gestureMgr.makeVisible(c.$$elmt);
		},
		/**
		 * Adds a panel to left or right sections - sidePanel class
		 */
		panelAttach: function(where, panel, action) {
			return this._panelsManager ? this._panelsManager.panelAttach(where, panel, action) : null;
		},
		/**
		 * Returns a panel - Just to have ctrlFactory declared here
		 */
		panelCreate: function(panelId, options) {
			return this._panelsManager ? this._panelsManager.panelCreate(panelId, options) : null;
		},
		/**
		 * Toggle left/right panel - data-action="sidePanel"
		 * 	panel : left or right
		 * 	action toggle
		 */
		_actToggleSidePanel: function(action, side) {
			if (!this._panelsManager) return;
			var self = this;
			var side = side == null || side === "all" ? ["left", "right"] : [side];
			side.forEach(function(where) {
				self._panelsManager.panelToggle(where, null, action);
			});
		},

		getPinPageId: function() {
			// Overridden by child classes!
		},
		getPinPageData: function() {
			var self = this;
			var pageData = {
				name: self.state.name,
				title: self.prototype.data("$title"),
				id: self.getPinPageId(),
				icon: "star"
			};
			return pageData;
		},
		/**
		 * Returns page authoring property
		 * path: string with or without '.' separator or array of string
		 */
		getAuthoring: function(path, defValue) {
			var val = utils.getPropByPath(this.article, path);
			return val != null ? val : defValue;
		}
	}
);

exports.Page = _Page;
