// syracuse-tablet/html/js/pages/pageDrafts.js
"use strict";

var globals = require('syracuse-tablet/html/js/helpers/globals');
var utils = require('syracuse-tablet/html/js/helpers/utils');
var uiRect = require('syracuse-tablet/html/js/ui/rect');
var Base = require('syracuse-tablet/html/js/pages/pageHtml').Page;
var locale = require('syracuse-tablet/html/js/helpers/locale');
var eventListener = require('syracuse-tablet/html/js/application/eventListener');
var factory = require('syracuse-tablet/html/js/application/appFactory');
var dispatcher = require('syracuse-tablet/html/js/sdata/sdataDispatcher');
var notifications = require('syracuse-tablet/html/js/helpers/notifications');
var fontUtils = require('syracuse-tablet/html/js/ui/fontUtils');
var HammerScroller = require('syracuse-tablet/html/js/helpers/hammerScroller').Klass;
var formatApi = require('syracuse-tablet/html/js/helpers/formatApi');


var _templates = {
	record: '\
			<section id="{{id}}" data-action="selectRecord" class="list-group-item s-m-draft-item {{status}}">\
				<div>\
					<div>\
						<div class="draft-title">{{title}}</div>\
						<div class="draft-date">{{creation_date}}</div>\
					</div>\
					<div>\
						<div class="draft-endpoint">{{endpoint}}</div>\
						<div class="draft-reason"><b>{{reason}}</b></div>\
					</div>\
					<div>\
						<div class="draft-comment"><i class="fa fa-comment-o"></i>{{comment}}</div>\
						<div class="draft-actions">\
							<span data-action="editDraft" class="fa fa-edit"></span>\
						</div>\
					</div>\
				</div>\
				<span class="badge checked fa fa-check">&nbsp;</span>\
			</section>',
	content: '\
			<div class="s-m-draft-header" {{#if isEmpty}}style="display:none"{{/if}}>\
				<div>\
					{{{selectMenu}}}\
				</div>\
				<div>\
					{{{filtersMenu}}}\
				</div>\
			</div>\
			<div id="draftInfo" {{#if isEmpty}}style="display:none"{{/if}}>{{info}}</div>\
			<div id="draftPageWrapper" class="s-m-scroll-wrapper">\
				<div class="list-group s-m-draft-list {{#if isEmpty}}empty{{/if}}">\
					{{{content}}}\
					<div id="draftSpinner">\
				   		<div>\
							<div class="spinner-loader"></div>\
						</div>\
					</div>\
				</div>\
			</div>\
		</div>',
	selectMenu: '\
		<div id="draftSelectMenuId" class="btn-group">\
			<div class="{{dropDirection}}">\
				<button type="button" class="btn btn-primary dropdown-toggle" data-toggle="dropdown" aria-expanded="false">\
					<span class="badge"></span>{{title}}<span class="caret"></span>\
				</button>\
				<ul class="dropdown-menu dropdown-menu-left" role="menu">\
					{{#each links}}\
						<li><a  {{#if $isHidden}}style="display:none"{{/if}} draggable="false"  data-action="selectMenu" data-params="{{params}}" href="#">{{title}}</a></li>\
					{{/each}}\
				</ul>\
			</div>\
		</div>',
	filtersMenu: '\
		<select id="draftFiltersMenu" data-action="filter">\
		{{#each filterOptions}}\
			<option value="{{value}}" {{#if selected}}selected{{/if}}>{{title}}</option>\
		{{/each}}\
		<select>'
};
//http://www.css-spinners.com/	
var _getHtml = function(name, ctx) {
	var key = name + "compiled";
	var tmpl = _templates[key];
	if (!tmpl) {
		tmpl = _templates[name];
		tmpl = _templates[key] = Handlebars.compile(tmpl);
	}
	return tmpl(ctx);
};

var _selectOptionsList = ["unselectall", "selectall", "hideunselected", "showunselected"];
var _toolbarLinks = [{
	name: "stop",
	$isHidden: true
}, {
	name: "synch",
	$isDisabled: true,
	$confirm: true
}, {
	name: "remove",
	$isDisabled: true,
	$confirm: true
}];

var _Page = utils.defineClass(
	function DraftsPage($parent, state, options) {
		options.header = true;
		options.breadcrumbs = true;
		options.footer = true;
		Base.call(this, $parent, state, options);
		this._selectedFilter = "creation_date:dsc";
		this._initLabels();
		notifications.subscribe(this, ["sm.action.link"]);
		this.dateFormatter = formatApi.getFormatter("application/x-datetime");
	}, Base, {
		scrollAllowed: function() {
			return false;
		},
		_initLabels: function() {
			this._filterOptions = [{
				value: "creation_date:asc",
				title: locale.text("drafts.filters.dateasc")
			}, {
				value: "creation_date:dsc",
				title: locale.text("drafts.filters.datedsc")
			}];
		},
		destroy: function() {
			Base.prototype.destroy.call(this);
			if (this._gestureMgr) {
				this._gestureMgr.destroy();
				this._gestureMgr = null;
			}
		},
		appendHtml: function($$parent, loadOpts) {
			var self = this;
			self.$$elmt.addClass("drafts");
			self._buildList().then(function(ok) {
				if (ok !== true) {
					$$parent.html("<br><br><b>" + locale.text("error") + "</b><br><br><b>" + locale.text("contact.admin") + "</b>");
					return;
				}
				self.$$myScrollWrapper = self.$$elmt.find("#draftPageWrapper");
				self.$$myScrollWrapper.css({
					height: "auto"
				});
				self._gestureMgr = new HammerScroller(self._$$list, {
					isPageScroller: true,
					direction: "v"
				});
			});
		},
		/**
		 * Resolve with true if ok
		 */
		_buildList: function() {
			var self = this;
			self.$$htmlRoot.empty();
			var deferred = $.Deferred();
			globals.getStorage().draftOperation("draftReadList").then(function(resources) {
				resources = self._applyFilters(resources || []);
				var content;
				if (resources.length > 0) {
					var recordsHtml = [];
					resources.forEach(function(rsrc) {
						recordsHtml.push(self._getRecordHtml(rsrc));
					});
					content = recordsHtml.join('\n');
				} else {
					content = locale.text("drafts.emptyList");
				}
				self.$$htmlRoot.html(_getHtml("content", {
					"content": content,
					"isEmpty": resources.length == 0,
					"filtersMenu": self._buildSortMenu(),
					"selectMenu": self._buildSelectMenu(),
					"info": locale.text("drafts.info", [resources.length])
				}));
				self._$$spinner = self.$$htmlRoot.find("#draftSpinner");
				self._$$list = self.$$htmlRoot.find(".s-m-draft-list");
				self._$$selectMenu = self.$$htmlRoot.find("#draftSelectMenuId");
			}).then(function() {
				self._updateSelectMenu();
				deferred.resolve(true);
			}).fail(function(e) {
				globals.getModal().error(locale.text("drafts.error.loading"), e, function() {
					deferred.resolve(false);
				});
			});
			return deferred.promise();
		},
		onMainPageResize: function(info, orientation, deviceClass) {
			var self = this;
			if (self.destroyed) return;
			Base.prototype.onMainPageResize.call(self, info, orientation, deviceClass);
			// handle gesture (necessary for small screen devices)
			setTimeout(function() {
				if (!self._gestureMgr) return;
				var mainRect = uiRect.elmtRect(self.$$contentElmt, "outer");
				var viewRect = uiRect.elmtRect(self.$$myScrollWrapper, "outer");
				var scrollRect = mainRect.intersectRect(viewRect);
				if (scrollRect.height < self._gestureMgr.$$elmt.height()) {
					self._gestureMgr.init(scrollRect, null, null, info && info.preserveScroll);
				} else {
					self._gestureMgr.reset();
				}
			});
		},
		/**
		 * Returns the html of a record
		 */
		_getRecordHtml: function(rsrc) {
			rsrc = $.extend({}, rsrc);
			var date = utils.isoDateTimeToDate(rsrc["creation_date"]).toISOString();
			rsrc["creation_date"] = this.dateFormatter.formatValue(date);
			rsrc["reason"] = locale.text("drafts.status." + rsrc.status);
			return _getHtml("record", rsrc);
		},
		_buildSelectMenu: function() {
			var links = [];
			_selectOptionsList.forEach(function(id) {
				links.push({
					title: locale.text("multiselect.action." + id),
					params: id,
					$isHidden: false
				});
			});
			return _getHtml("selectMenu", {
				links: links,
				title: locale.text("multiselect.selected"),
				dropDirection: "dropdown"
			});
		},
		_buildSortMenu: function() {
			var self = this;
			self._selectedFilter = self._selectedFilter || "creation_date:asc";
			var opts = [];
			self._filterOptions.forEach(function(item) {
				item = $.extend({}, item);
				item.selected = item.value === self._selectedFilter;
				opts.push(item);
			});
			return _getHtml("filtersMenu", {
				"filterOptions": opts
			});
		},
		/**
		 * Update select menu according to the status of the records
		 * Update eth number of selected records
		 */
		_updateSelectMenu: function() {
			var allRecords = this._selectRecords("all").length;
			var selectedRecords = this._selectRecords(".selected").length;
			this._$$selectMenu.find('a[data-action]').toggle(selectedRecords > 0);
			this._$$selectMenu.find('a[data-params="selectall"]').toggle(allRecords !== selectedRecords);
			var hiddenUnselRecords = this._selectRecords(":not(.selected):not(:visible)").length;
			this._$$selectMenu.find('a[data-params="hideunselected"]').toggle(selectedRecords > 0 && selectedRecords < allRecords && hiddenUnselRecords == 0);
			this._$$selectMenu.find('a[data-params="showunselected"]').toggle(selectedRecords > 0 && hiddenUnselRecords > 0);
			this._$$selectMenu.find('.badge').text(selectedRecords);
			this.$$elmt.find(".draft-synch, .draft-remove").toggleClass("disabled", selectedRecords == 0);
		},
		/**
		 * Select records accordinbt to cssSelector
		 */
		_selectRecords: function(cssSelector) {
			var selector = "";
			if (cssSelector != "all") {
				selector += cssSelector;
			}
			return this._$$list.find("section.s-m-draft-item" + selector);
		},
		_getRecordId: function($$e) {
			return $$e.closest("section.s-m-draft-item").attr("id");
		},
		/**
		 * Edit draft - Switch endpoint if needed
		 */
		_actEditDraft: function(params, $$e, evt) {
			if (this._busy === true) {
				return;
			}
			// Get the id and triggers application.openDraft
			eventListener.triggerAction(this.$$elmt, "openDraft", this._getRecordId($$e));
		},
		/**
		 * Filter records
		 */
		_actFilter: function() {
			if (this._busy === true) {
				return;
			}
			var val = $(arguments[1]).val();
			if (val == this._selectedFilter) {
				return;
			}
			this._selectedFilter = val;
			this._buildList();
		},
		_applyFilters: function(resources) {
			if (!this._selectedFilter) {
				return resources;
			}
			var f = this._selectedFilter.split(":");
			if (f.length != 2) {
				return resources;
			}
			var property = f[0];
			var order = f[1];
			resources.sort(function(a, b) {
				a = a[property];
				b = b[property];
				if (a == b) return 0;
				if (a > b) return order == "asc" ? +1 : -1;
				if (a < b) return order == "asc" ? -1 : +1;
			});
			return resources;
		},
		/**
		 * Secltion/Unselection of records
		 */
		_actSelectMenu: function(action) {
			var self = this;
			switch (action) {
				case "selectall":
				case "unselectall":
					// First show all
					var elmts = self._selectRecords("all").show();
					if (action == "unselectall") {
						elmts = self._selectRecords(".selected");
					}
					elmts.each(function() {
						self._toggleRecordSelection($(this), action === "selectall");
					});
					break;
				case "hideunselected":
					self._selectRecords(":not(.selected)").hide();
					break;
				case "showunselected":
					self._selectRecords(":not(.selected)").show();
					break;
			}
			this._updateSelectMenu();
		},
		/**
		 * Toggle record's selection
		 */
		_actSelectRecord: function(params, $$elmt, force) {
			var $$rec = $$elmt.closest(".s-m-draft-item");
			this._toggleRecordSelection($$elmt.closest(".s-m-draft-item"));
			this._updateSelectMenu();
		},
		/**
		 * force true/false/undefined
		 */
		_toggleRecordSelection: function($$rec, force) {
			if (!$$rec || $$rec.length == 0) return;
			$$rec.toggleClass("selected", force);
		},
		/**
		 * Start upload/remove
		 */
		_mainProcessBegin: function(name, ids) {
			this._stop = false;
			this._busy = true;
			this._logData = {
				nbSuccess: 0,
				nbErrors: 0,
				log: {}
			};
			this.$$elmt.find(".draft-stop").removeClass("disabled").show();
			this._selectRecords(":not(.selected)").hide();
		},
		/**
		 * End upload/remove
		 */
		_mainProcessEnd: function(name) {
			this._busy = false;
			this.$$elmt.find(".draft-stop").hide();
			this._updateSelectMenu();
			this.$$elmt.find(".synchronized").removeClass("synchronized");
			// Update draft counter
			this.$$elmt.find("#draftInfo").text(locale.text("drafts.info", [this._selectRecords("all").length]));
			this._selectRecords(":not(.selected)").show();
		},
		/**
		 * Upload alld the records given by ids array and resolve like _uploadPop
		 * Resolves stoppewith stopped/completed
		 */
		_upload: function(ids) {
			var self = this;
			self._mainProcessBegin("upload", ids);
			var join = $.Deferred();
			self._uploadPop(ids).then(function(status) {
				self._mainProcessEnd("upload");
				join.resolve(status);
			}).fail(function(e) {
				self._mainProcessEnd("upload", ids);
				join.reject(e);
			});
			return join.promise();
		},
		/**
		 * Pop the ids ids array and upload them
		 * Returns the promise that will be resolve when:
		 * 	- upload succeeded (status = completed)
		 * 	- upload failed with an exception
		 * 	- upload stopped (status = stopped)
		 * _mainDeferred is null when we call _uploadPop outside the method
		 */
		_uploadPop: function(ids, _mainDeferred) {
			var first = _mainDeferred == null;
			if (first) {
				_mainDeferred = $.Deferred();
			}
			try {
				if (this._stop === true) {
					_mainDeferred.resolve("stopped");
					return;
				}
				if (!ids || ids.length == 0) {
					_mainDeferred.resolve("completed");
					return;
				}
				var self = this;
				self._uploadOne(ids.pop()).then(function() {
					setTimeout(function() {
						self._uploadPop(ids, _mainDeferred);
					});
				}).fail(function(e) {
					_mainDeferred.reject(e);
				});
			} catch (e) {
				_mainDeferred.reject(e);
			}
			return first ? _mainDeferred.promise() : null;
		},
		/**
		 * Upload the record given by id
		 * Resolve if no exception
		 * If upload succeeds 		-> Remove the draft and resolve
		 * If upload not succeeds 	-> Update the draft with new error and resolve
		 * If exception --> Reject
		 */
		_uploadOne: function(id) {
			var self = this;
			var deferred = $.Deferred();
			self._uploadOneBegin(id).then(function(ctx) {
				if (!ctx || !ctx.restoredContext) {
					throw new Error("unexpected null draft context");
				}
				if (!ctx.saveLink || !ctx.saveLink.sDataUrl) {
					throw new Error("unexpected null save link info");
				}
				return globals.getMetaData().getPageDetails(ctx.restoredContext.currentState).then(function(info) {
					return $.smResolve(ctx, info);
				}).fail(function(e) {
					return $.smReject(e);
				});
			}).then(function(draftCtx, info) {
				var dao = factory.createDaoSdata(draftCtx.dataSet, info.prototype);
				var payload = dao.getActionPayload(draftCtx.saveLink, true);
				return dispatcher.dispatch({
					$url: draftCtx.saveLink.sDataUrl,
					$method: draftCtx.saveLink.$method
				}, payload).then(function(result) {
					if (!result) {
						throw new Error("unexpected null http response");
					}
					var responseJSON = result.responseJSON || {};
					var next;
					if (!result.isSuccess) {
						if (result.status === 401) {
							// KO - Reject to display a dialog and stop process
							next = $.smReject(locale.text("drafts.error.unauthorized"));
						} else if (result.responseJSON && result.responseJSON.$noConnectionError === true) {
							// KO - Reject to display a dialog and stop process
							next = $.smReject(locale.text("drafts.error.noconnection"));
						} else {
							// KO - Set up draft with new error and date
							dao.updateMeta(responseJSON);
							draftCtx.status = "error";
							draftCtx["creation_date"] = utils.getCurISODateTime(new Date());
							next = self._uploadOneEnd(id, "saveFailed", {
								draftCtx: draftCtx
							});
						}
					} else {
						// OK
						next = self._uploadOneEnd(id, "saveSucceeded");
					}
					return next;
				});
			}).then(function() {
				deferred.resolve();
			}).fail(function(e) {
				if (typeof e === "string") {
					e = {
						$diagnoses: [{
							$message: e
						}]
					};
				}
				// Exception expected
				return self._uploadOneEnd(id, "exception", {
					exception: e
				}).always(function(e) {
					deferred.reject(e);
				});
			});
			return deferred.promise();
		},
		/**
		 * Set the ui context
		 * Reads the draft
		 */
		_uploadOneBegin: function(id) {
			var $$rec = this.$$elmt.find("#" + id);
			if (this._$$spinner) {
				var pos = $$rec.position();
				this._$$spinner.css({
					top: pos.top + "px",
					display: 'table'
				});
				this._$$spinner.children().outerHeight($$rec.outerHeight());
			}
			this.getGestureMgr().makeVisible($$rec);
			return globals.getStorage().draftOperation("draftRead", id);
		},
		/**
		 * Set ui context
		 * Save/Delete the draft according to the status
		 */
		_uploadOneEnd: function(id, status, opts) {
			var self = this;
			self._logData.log[id] = {
				status: status
			};
			if (this._$$spinner) {
				this._$$spinner.hide();
			}
			var $$record = this.$$elmt.find("#" + id);
			if (status === "saveSucceeded") {
				return globals.getStorage().draftOperation("draftDelete", id).then(function() {
					self._logData.nbSuccess++;
					$$record.remove();
				}).fail(function(e) {
					return $.smReject(e);
				});
			} else if (status === "saveFailed") {
				self._logData.nbErrors++;
				return globals.getStorage().draftOperation("draftSave", opts.draftCtx).then(function() {
					var $$newRecord = $(self._getRecordHtml(opts.draftCtx)).addClass("selected synchronized");
					$$record.after($$newRecord);
					$$record.remove();
				}).fail(function(e) {
					return $.smReject(e);
				});
			} else if (status === "exception") {
				return $.smReject(opts.exception);
			} else {
				// Exception
				return $.smReject(new Error("Unexpected draft status"));
			}
		},
		/**
		 * Upload all the drafts
		 */
		_uploadAll: function() {
			var self = this;
			self._upload(self._selectRecords(".selected").map(function() {
				return this.id;
			}).get().reverse()).then(function(status) {
				var nbSuccess = self._logData.nbSuccess;
				var nbErrors = self._logData.nbErrors;
				var count = nbSuccess + nbErrors;
				globals.getModal().notify({
					title: locale.text("drafts.synch." + status),
					body: locale.text("drafts.synch.result", [count, nbSuccess, nbErrors]),
					severityClass: nbErrors > 0 ? "warning" : "success"
				});
			}).fail(function(e, status) {
				globals.getModal().error(locale.text("drafts.error.uploading"), e);
			});
		},
		_removeAll: function() {
			var self = this;
			globals.getStorage().draftOperation("draftDeleteAll").then(function() {
				self._buildList();
			}).fail(function(e) {
				globals.getModal().error(locale.text("drafts.error.removing"), e);
			});
		},
		/** Event sm.action.link
		 *  Notified when a action succeeded - sdata link - see action manager
		 *  success:	true/false
		 *  link:		Link info
		 *  result:		Json response
		 */
		notifActionLink: function(success, link, result, options) {
			if (!this.isActive()) return;
			switch (link.name) {
				case "synch":
					this._uploadAll();
					break;
				case "remove":
					this._removeAll();
					break;
				case "stop":
					this._stop = true;
					break;
			}
		},
		/**
		 * Returns the links to display in the header/footer
		 */
		getToolbarLinks: function(toolbar, options) {
			var device = globals.getSiteLayout().getDeviceType();
			if ((device === "tablet" && toolbar === "footer") || (device === "smartphone" && toolbar === "header")) {
				var result = [];
				_toolbarLinks.forEach(function(act) {
					result.push({
						"name": act.name,
						"title": locale.text("drafts.action." + act.name),
						"icon": fontUtils.getDraftIcons(act.name),
						"$uuid": utils.UUID(),
						"css": "draft-" + act.name,
						"footerHeader": true,
						"isClientAction": true,
						"$confirm": act.$confirm === true ? locale.text("drafts.dlog.confirm." + act.name) : null,
						"$isHidden": act.$isHidden === true,
						"$isDisabled": act.$isDisabled === true
					});
				});
				return result;
			}
			return null;
		}
	});

exports.Page = _Page;
