// syracuse-tablet/html/js/pages/pageCardDesign.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var globals = require('syracuse-tablet/html/js/helpers/globals');
var Base = require('syracuse-tablet/html/js/pages/pageRegular').Page;
var daoSdata = require('syracuse-tablet/html/js/application/daoSdata');

/**
 * Child page that displays card detail for design
 */
var _Page = utils.defineClass(
	function CardDesignPage($parent, state, prototype, article, parentPage, options) {
		options.header = false;
		options.footer = false;
		options.breadcrumbs = false;
		if (!parentPage) throw new Error("carddesign page - Unexpected empty parent page");
		this.parentPage = parentPage;
		Base.call(this, $parent, state, prototype, article, options);
	}, Base, {

		_initSdataInfo: function(sdataUrl, sdataMethod) {
			// nothing no sdataUrl
		},

		destroy: function() {
			if (this.destoyed) return;
			Base.prototype.destroy.call(this);
			this.parentPage = null;
			this._notifData = null;
			this._row = null;
		},
		buildHtmlOption: function(loadOptions) {
			return Base.prototype.buildHtmlOption.call(this, loadOptions).
			then(function(opts) {
				opts.toolBars.showHome = true;
				opts.cardDesign = true;
				return opts;
			});
		},
		isEditMode: function() {
			return false;
		},

		loadData: function(loadOptions) {
			// We get the row DAO - !!! Do not store parentPage
			var ctrl = this.parentPage.getControl(this.state.options.controlId);
			if (!ctrl) throw new Error("cardDesign page - Array datasource not found");
			var dao;
			if (ctrl.getArrayData() && ctrl.getArrayData().$resources.length > 0) {
				if (!this._row) {
					// We store to keep the same record when we change an authoring property
					this._row = ctrl.getRowById(ctrl.$$elmt.find(".s-m-record.s-m-auth-rec-selected").attr("data-params"));
					this._row = this._row ? this._row.row : ctrl.getArrayData().$resources[0];
					this._row = $.extend(true, {}, this._row.data);
				}
				dao = new daoSdata.Dao(this._row, this.prototype, null, "carddesign");
			} else {
				dao = daoSdata.getArrayEmptyRow(this, this.prototype, "carddesign", ctrl.getDao());
			}
			this.setDao(dao);
			return $.smResolve();
		},
		/**
		 * Return data for parent
		 */
		getParentNotif: function() {
			if (this._notifData) {
				return {
					parentId: this.state.options.parentId,
					controlId: this.state.options.controlId,
					notifId: "sm.updt.card.article",
					notifData: this._notifData
				};
			}
			return null;
		},
		_actSaveCardDesign: function() {
			this._notifData = {};
			this._notifData.article = $.extend(true, {}, this.article);
			globals.getApplication().goBack();
		}
	});

exports.Page = _Page;
