// syracuse-tablet/html/js/pages/pageCardDesign.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var globals = require('syracuse-tablet/html/js/helpers/globals');
var Base = require('syracuse-tablet/html/js/pages/pageRegular').Page;
var daoSdata = require('syracuse-tablet/html/js/application/daoSdata');

/**
 * Child page that displays card detail for design
 */
var _Page = utils.defineClass(
	function($parent, state, prototype, article, parentPage, options) {
		if (!parentPage) throw new Error("carddesign page - Unexpected empty parent page");
		this.parentPage = parentPage;
		Base.call(this, $parent, state, prototype, article, options);
	}, Base, {

		_initSdataInfo: function(sdataUrl, sdataMethod) {
			// nothing no sdataUrl
		},

		destroy: function() {
			Base.prototype.destroy.call(this);
			this.parentPage = null;
			this._notifData = null;
		},

		isEditMode: function() {
			return false;
		},

		loadData: function() {
			// We get the row DAO - !!! Do not store parentPage
			var ctrl = this.parentPage.getControl(this.state.options.controlId);
			if (!ctrl) throw new Error("rowdetail page - Array datasource not found");
			var dao;
			if (ctrl.getArrayData() && ctrl.getArrayData().$resources.length > 0) {
				var data = $.extend(true, {}, ctrl.getArrayData().$resources[0].data);
				dao = new daoSdata.Dao(data, this.prototype, null, "carddesign");
			} else {
				dao = daoSdata.getArrayEmptyRow(this, this.prototype, "carddesign", ctrl.getDao());
			}
			this.setDao(dao);
			return $.smResolve();
		},
		/**
		 * Return data for parent
		 */
		getParentNotif: function() {
			if (this._notifData) {
				return {
					parentId: this.state.options.parentId,
					controlId: this.state.options.controlId,
					notifId: "sm.updt.card.article",
					notifData: this._notifData
				};
			}
			return null;
		},
		_actSaveCardDesign: function() {
			this._notifData = {};
			this._notifData.article = $.extend(true, {}, this.article);
			globals.getApplication().goBack();
		}
	});

exports.Page = _Page;
