// syracuse-tablet/html/js/pages/pageLookup.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var Base = require('syracuse-tablet/html/js/pages/pageRegular').Page;
var eventListener = require('syracuse-tablet/html/js/application/eventListener');

/**
 * Lookup page that handles selection
 * TODO - eventually if we have to merge wit regular page later
 */
var _Page = utils.defineClass(
	function($parent, state, prototype, article) {
		Base.call(this, $parent, state, prototype, article);
		this.selectedRowJson = null;
	}, Base, {

		addPanel: function(panelId) {
			if (panelId == "footer" || panelId == "rightpanel" || panelId == "leftpanel") return null;
			return Base.prototype.addPanel.call(this, panelId);
		},

		destroy: function() {
			Base.prototype.destroy.call(this);
			this.selectedRowJson = null;
		},
		/**
		 * Do not call super (Page) because we just need to store the rowDao and close the page
		 * Data selected will be passed to parent page by pageLoader via standard parent notification process - getParentNotif
		 */
		notifSelectRow: function(arrayId, rowId, rowDao) {
			if (!rowDao) return;
			// Clone JSON because array dao is destroyed with the page
			this.selectedRowJson = $.extend(true, {}, rowDao.data);
			if (this.isVignette) {
				eventListener.triggerAction(this.$$elmt, "vignetteBack");
			} else {
				this.goBack();
			}
		},
		/**
		 *  Modify title for dvlp tests
		 */
		buildHtmlOption: function() {
			var opts = Base.prototype.buildHtmlOption.call(this);
			opts.header.title = "Lookup page";
			return opts;
		},
		/**
		 * getParentNotif method is called by pageLoader to notify parent page
		 */
		getParentNotif: function() {
			if (!this.selectedRowJson) return null;
			return {
				parentId: this.state.options.parentId,
				controlId: this.state.options.controlId,
				notifId: "sm.lookup.selection",
				notifData: this.selectedRowJson
			};
		},

	});

exports.Page = _Page;
