// syracuse-tablet/html/js/pages/pageWelcomeDashboard.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var Base = require('syracuse-tablet/html/js/pages/pageDashboard').Page;
var locale = require('syracuse-tablet/html/js/helpers/locale');

var _Page = utils.defineClass(

	function WelcomeDashboardPage($parent, state, prototype, article) {
		Base.call(this, $parent, state, prototype, article);
		this.vignettes = [];
	}, Base, {

		destroy: function() {
			Base.prototype.destroy.call(this);
		},

		loadData: function() {
			return null;
		},

		beforeAddContent: function() {
			this.addTitle(this.prototype.data("$title"), this.prototype.data("$description"));
		},

		addControl: function(c, isPageControl) {
			Base.prototype.addControl.call(this, c, isPageControl);
			if (c && c.isVignette()) {
				this.vignettes.push(c);
			}
		},
		afterRender: function(options) {
			var self = this;
			Base.prototype.afterRender.call(this, options);
			// Empty dashboard
			if (!this.vignettes || this.vignettes.length < 1) {
				this.renderEmptyDashboardMessage();
			} else {
				this.vignettes.forEach(function(v) {
					var options = {
						sdataParameters: self.state.options && self.state.options.sdataParameters || {}
					};
					v.load(null, options);
				});
			}
		},
		buildHtmlOption: function() {
			var self = this;
			var opts = Base.prototype.buildHtmlOption.call(self);
			opts.makeFullHeight = false;
			opts.toolBars = {
				showAuthoring: false,
				showLinks: true,
				showHome: false,
				statusPinPage: 0,
			};
			return opts;
		},
	});

exports.Page = _Page;
