// msoffice/lib/excel/html/config.js

if (!require('streamline/module')(module)) 
    return;

window.getStack = function(){
    try {
        var stack = [];
        var f = arguments.callee.caller;
        while (f) {
            var fn = f.toString();
            var fname = fn.substring(fn.indexOf('function') + 8, fn.indexOf('(')) || 'anonymous';
            stack.push(fname);
            f = f.caller;
        }
        return stack.join('\n');
    } 
    catch (e) {
    }
}

window.onerror = function(message, url, line){
    alert(message + '\nurl=' + url + '\nat line ' + line + '\n' + window.getStack());
    return false;
}

var siteFactory = require('syracuse-ui/lib/office/excel/officeConfigSite');
exports.main = function(){
    siteFactory.load({
        $layoutOptions: {
            page: {
                dashboard: {
                    $skin: "s-dashboard",
                    section: "s-dashboard-h1",
                    block: "s-dashboard-h2",
                    menus: "s-h1-menus"
                },
                portlet: {
                    $skin: "s-portlet",
                    section: "s-portlet-h1",
                    block: "s-portlet-h2",
                    menus: "s-h2-menus"
                },
                page: {
                    $skin: "s-page",
                    section: "s-page-h1",
                    block: "s-page-h2",
                    menus: "s-h1-menus"
                },
                fusionpage: {
                    $skin: "s-page",
                    section: "s-page-h1",
                    block: "s-page-h2",
                    menus: "s-h1-menus"
                }
            },
            searchFacet: {
                menus: "s-list-menus"
            },
            authoring: {
                $skin: "s-author-designer-item",
                section: "s-author-designer-item-h1",
                block: "s-author-designer-item-h2",
                menus: "s-h1-menus"
            },
            list: {
                menus: "s-list-menus",
                
                builder: {
                    grid: {
                        section: "s-page-h2",
                        block: "s-page-h2",
                        recordMenus: "s-list-record-menus"
                    },
                    vgrid: {
                        section: "s-page-h2",
                        block: "s-page-h2",
                        recordMenus: "s-list-record-menus"
                    },
                    cardList: {
                        section: "s-page-h2",
                        block: "s-page-h2",
                        recordMenus: "s-list-record-menus"
                    },
                    singleField: {},
                    reference: {
                        recordMenus: "s-list-record-menus"
                    }
                }
            
            }
        
        },
		$iconPath: "/syracuse-ui/themes/desktop/images-general/",
        $module: "X3 for Office",
        $pageTemplateUrl: "/sdata/syracuse/collaboration/syracuse/pages('{endpoint}.{representation},{target}')?role={$role}",
        $userProfileUrl: "/sdata/syracuse/collaboration/syracuse/userProfiles/$template/$workingCopies?representation=userProfile.$edit&role={$role}",
        $syracuseMainPageUrl: "/syracuse-main/html/main.html"
    }, {
        $links: {
            $home: {
                $title: "Home",
                $url: "?representation=excelconfig.$dashboard"
            }
        },
        $navigation: {
            $about: {
                "aboutus": {
                    "$title": "About us"
                },
                "privacypolicy": {
                    "$title": "Privacy Policy"
                },
                $help: {
                    $title: "Help",
                    $url: "http://www.sage.fr/mge/logiciels-erp/sage-erp-x3-premium-edition",
                    $target: "blank"
                },
                $contactus: {
                    "$title": "Contact us",
                    $url: "http://www.sage.com/contactus",
                    $target: "blank"
                }
            }
        },
        $: {}
    });
};

if (require.main == module) 
    exports.main();
