// msoffice/lib/excel/html/main.js

if (!require('streamline/module')(module)) 
    return;

window.getStack = function(){
    var stack = [];
    var f = arguments.callee.caller;
    while (f) {
        var fn = f.toString();
        var fname = fn.substring(fn.indexOf('function') + 8, fn.indexOf('(')) || 'anonymous';
        stack.push(fname);
        f = f.caller;
    }
    return stack.join('\n');
}

window.onerror = function(message, url, line){
    alert(message + '\nurl=' + url + '\nat line ' + line + '\n' + window.getStack());
    return false;
}

var siteFactory = require('syracuse-ui/lib/office/excel/officeSite');
exports.main = function(){
    siteFactory.load({
        $layoutOptions: {
            page: {
                dashboard: {
                    $skin: "s-dash",
                    section: "s-dash-h1",
                    block: "s-dash-h2",
                    menus: "s-page-menus"
                },
                vignette: {
                    $skin: "s-vignette",
                    section: "s-h3",
                    block: "s-h4",
                    menus: "s-vignette-menus"
                },
                page: {
                    $skin: "s-page",
                    section: "s-h1",
                    block: "s-h2",
                    menus: "s-page-menus"
                }
            },
            searchFacet: {
                menus: "s-list-menus"
            },
            authoring: {
                $skin: "s-aw-designer-item",
                section: "s-aw-designer-h1",
                block: "s-aw-designer-h2",
                menus: "s-page-menus"
            },
            list: {
                menus: "s-list-menus",
                
                builder: {
                    grid: {
                        section: "s-h2",
                        block: "s-h2",
                        recordMenus: "s-record-menus"
                    },
                    vgrid: {
                        section: "s-h2",
                        block: "s-h2",
                        recordMenus: "s-record-menus"
                    },
                    cards: {
                        section: "s-h2",
                        block: "s-h2",
                        recordMenus: "s-record-menus"
                    },
                    singleField: {},
                    reference: {
                        recordMenus: "s-record-menus"
                    }
                }
            
            }
        
        },
        $iconPath: "/syracuse-ui/themes/desktop/images-general/",
        $module: "X3 for Office",
        $pageTemplateUrl: "/sdata/syracuse/collaboration/syracuse/pages('{endpoint}.{representation},{target}')?role={$role}",
        $userProfileUrl: "/sdata/syracuse/collaboration/syracuse/userProfiles/$template/$workingCopies?representation=userProfile.$edit&role={$role}",
        $syracuseMainPageUrl: "/syracuse-main/html/main.html",
        $isUserProfileQuickViewEnabled: false
    }, {
        $links: {
            $home: {
                $title: "Home",
                $url: "?representation=excelhome.$dashboard"
            }
        },
        $properties: {}
    });
};

if (require.main == module) 
    exports.main();
