// syracuse-ui/lib/page/delta/deltaManager.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function DeltaManager() {}

exports.DeltaManager = helpers.defineClass(DeltaManager, null, {
	applyObjectArrayDelta: function(targetList, sourceList) {
		if (sourceList) {
			if (targetList.length == sourceList.length) {
				for (var ii = 0, jj = sourceList.length; ii < jj; ii++) {
					this.applyObjectDelta(targetList[ii], sourceList[ii]);
				}
			} else {
				targetList.splice(0, targetList.length);
				for (var ii = 0, jj = sourceList.length; ii < jj; ii++) {
					targetList.push(sourceList[ii]);
				}
			}
		}
		return targetList;
	},
	isVariantArray: function(list) {
		if (list.length) {
			if (list[0].$uuid == undefined) {
				if (typeof(list[0]) == 'object' && Object.keys(list[0]).length == 1) {
					return true;
				}
			}
		}
		return false;
	},
	applyPageArrayDelta: function(targetList, sourceList) {
		if (sourceList) {
			var targetMap = {};
			var isVariant = this.isVariantArray(sourceList);
			if (this.page.$isPartialDelta || (sourceList.length == 1 && sourceList[0].$index !== undefined)) {
				var found, foundIndex;
				for (var ii = 0, jj = sourceList.length; ii < jj; ii++) {
					found = null;
					var sourceRecord = sourceList[ii];
					foundIndex = 0;
					for (var mm = targetList.length; foundIndex < mm; foundIndex++) {
						found = targetList[foundIndex];
						if (found.$uuid == sourceRecord.$uuid) {
							break;
						} else {
							found = null;
						}
					}
					if (found) {
						if (sourceRecord.$isDeleted) {
							targetList.splice(foundIndex, 1);
						} else {
							this.applyObjectDelta(found, sourceRecord);
							if (sourceRecord.$index !== undefined && sourceRecord.$index != foundIndex) {
								targetList.splice(foundIndex, 1);
								targetList.splice(sourceRecord.$index, 0, found);
								delete found.$index;
							}
						}
					} else {
						if (!sourceRecord.$isDeleted) {
							if (sourceRecord.$index !== undefined) {
								targetList.splice(sourceRecord.$index, 0, sourceRecord);
							} else {
								targetList.push(sourceRecord);
							}
						}
					}
				}
			} else {
				for (var ii = 0, jj = targetList.length; ii < jj; ii++) {
					var record = targetList[ii];
					if (record) {
						if (record.$uuid !== undefined) {
							targetMap[record.$uuid] = record;
						} else {
							if (isVariant) {
								record = record[Object.keys(record)[0]];
								if (record.$uuid !== undefined) {
									targetMap[record.$uuid] = record;
								}
							}
						}

					}
				}
				targetList = [];
				for (var ii = 0, jj = sourceList.length; ii < jj; ii++) {
					var record = sourceList[ii];
					if (isVariant) {
						var variantKey = Object.keys(record)[0];
						this._applyPageRecordDelta(record, record[variantKey], targetMap, targetList, variantKey);
					} else {
						this._applyPageRecordDelta(record, record, targetMap, targetList);
					}

				}
			}
		}
		return targetList;
	},
	//if variantKey is not defined record  == subRecord
	_applyPageRecordDelta: function(record, subRecord, targetMap, targetList, variantKey) {
		if (subRecord.$uuid !== undefined) {
			if (!subRecord.$isDeleted) {
				if ((subRecord.$uuid !== undefined) && targetMap[subRecord.$uuid]) {
					if (variantKey) {
						var newNecord = this.applyObjectDelta(targetMap[subRecord.$uuid], subRecord);
						record = {};
						record[variantKey] = newNecord;
						targetList.push(record);
					}
				} else {
					targetList.push(this.applyObjectDelta(targetMap[subRecord.$uuid], subRecord));
				}
			}
		} else {
			targetList.push(record);
		}
	},
	ensureServerIndex: function(list) {
		if (this.isVariantArray(list)) {
			for (var ii = 0, jj = list.length; ii < jj; ii++) {
				var record = list[ii];
				record[Object.keys(record)[0]].$serverIndex = ii;
			}
		} else {
			for (var ii = 0, jj = list.length; ii < jj; ii++) {
				list[ii].$serverIndex = ii;
			}
		}

	},
	applyObjectDelta: function(target, source, isObjectDelta) {
		if (source) {
			var properties = Object.keys(source);
			for (var ii = 0, jj = properties.length; ii < jj; ii++) {
				var property = properties[ii];
				var targetValue = target[property];
				var sourceValue = source[property];
				if (typeof(sourceValue) == 'object') {
					if (!(targetValue == null || sourceValue === null)) {
						if (Array.isArray(sourceValue)) {
							if (isObjectDelta) {
								sourceValue = this.applyObjectArrayDelta(targetValue, sourceValue);
							} else {
								if (this.page.$isEditMode || this.page.$isPartialDelta) {
									sourceValue = this.applyPageArrayDelta(targetValue, sourceValue);
								}
							}
						} else {
							sourceValue = this.applyObjectDelta(targetValue, sourceValue, isObjectDelta);
						}
					}
					if (Array.isArray(sourceValue)) {
						this.ensureServerIndex(sourceValue);
					}
				}
				target[property] = sourceValue;
			}
		}
		return target;
	}
});
