// @sage/syracuse-lib/src/collaboration/representations/historyLog.ts
"use strict";
exports.representation = {
	resources: function() {
		var locale = require('streamline-locale');
		var globals = require('streamline-runtime').globals;
		return locale.resources(module, globals.context && globals.context.sessionLocale)();
	},
	$entityName: "historyLog",
	$facets: {
		$details: {
			$layout: {
				$items: [{
					"$category": "section",
					$layout: {
						$items: [{
							$bind: "$creDate"
						}, {
							$bind: "$creUser"
						}, {
							$bind: "name"
						}, {
							$bind: "value"
						}, {
							$bind: "op"
						}, {
							$bind: "items",
						}]
					}
				}]
			}
		},
		$edit: {
			$copy: "$details"
		},
		$query: {
			$layout: {
				$items: [{
					$category: "section",
					$layout: {
						$items: [{
							$category: "section",
							$layout: {
								$items: [{
									$bind: "$resources",
									$layout: {
										$items: [{
												$bind: "$creDate"
											}, {
												$bind: "$creUser"
											}, {
												$bind: "op"
											}, {
												$bind: "name"
											}, {
												$bind: "value"
											}

										]
									}
								}]
							}
						}]
					}
				}]
			}
		}
	}
};
