// syracuse-ui/lib/field/array/cardsBuilder/cardRecord.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var RecordArticle = require("syracuse-ui/lib/field/array/recordArticle").RecordArticle;

function CardRecord(){
}

exports.CardRecord = helpers.defineClass(CardRecord, RecordArticle, {
    initializeRecord: function(options){
        var rowIndex;
        if (this.builder.rows.length == 0) {
            this.row = this.builder.appendCardsRow();
        }
        else {
            this.row = this.builder.rows[this.builder.rows.length - 1];
        }
        if (this.row.count == this.list.$item.$cardsByRowCount) {
            this.row = this.builder.appendCardsRow();
        }
        this.cellIndex = this.row.count++;
        options.$item = {
            $category: "section",
            $layout: {
                $items: this.list.$item.$layout.$items
            }
        };
    },
    notifyReorder: function(){
        var self = this;
        var rowIndex = 0;
        self.builder.rows.forEach(function(row){
            row.count = 0;
            row.cells = [];
        });
        var rowOffset = 0;
        self.builder.list._store._records.forEach(function(record, index){
            record.row = self.builder.rows[rowOffset];
            record.cellIndex = record.row.count++;
            record.$$item.detach().appendTo($(record.row.slot));
            record.row.cells.push(record.$$item[0]);
            if (record.row.count >= record.row.max) {
                rowOffset++;
            }
        });
        var lastRow = self.builder.rows[self.builder.rows.length - 1];
        var $$lastRow = $(lastRow.slot);
        var $$children = $$lastRow.children(".s-cards-cell");
        if ($$children.length) {
            for (var ii = 0; ii < $$children.length; ii++) {
                lastRow.cells.push($($$children[0]).detach().appendTo($$lastRow)[0]);
            }
        }
    },
    remove: function(){
        this.$$item.remove();
    },
    drawBox: function(){
        var item = this.row.cells[this.cellIndex];
        item.setAttribute("data-s-record", this.$recordUuid);
        item.className = this.list.$recordCss;
        this.$$item = $(item);
        if (this.list.$item.$alternateStyle && this.$recordIndex % 2) {
            item.className += " " + this.list.$skin + "-alt";
        }
        var selectorSlot = item.appendChild(document.createElement("div"));
        if (this.list._store.selector.appendToRecord) {
            this._selectorCheck = this.list._store.selector.appendToRecord(selectorSlot);
        }
        else {
            selectorSlot.style.display = "none";
        }
        
        var toolbar = document.createElement("div");
        toolbar.className = this.list.$skin + "-toolbar";
        item.appendChild(toolbar)
        
        if (this.list.$item.$isRowIndexVisible) {
            var dom = document.createElement("div");
            dom.className = this.list.$skin + "-row-index";
            if (this.list.$capability.reorder) {
                dom.className += " s-list-record-reorder";
            }
            $(toolbar.appendChild(dom)).text((this.list._store.pagingCapability._$startIndex || 0) + this.$recordIndex);
        }
        if (!this.list.$item.$isMenuRecordHidden) {
            this._contextMenusSlot = document.createElement("div");
            this._contextMenusSlot.className = this.list.$skin + "-actions-cell";
            toolbar.appendChild(this._contextMenusSlot);
            this.list.applyCapabilities(this);
        }
        
        var card = item.appendChild(document.createElement("div"));
        card.className = this.list.$skin + "-record";
        if (this.list._store.selector.isRowMode) {
            card.className += " s-list-selector-row";
            card.setAttribute("data-s-record", this.$recordUuid);
        }
        
        document.itemFactory.load($(card), this.list.$item.$cards, this);
        item.style.visibility = "";
    },
    highlightSelection: function(selected){
        document.site.toggleClass(this.$$item[0], "s-list-record-selected", selected);
        if (this._selectorCheck) {
            $(this._selectorCheck).attr("checked", selected);
        }
    },
    dispose: function(){
        delete this.row;
        RecordArticle.prototype.dispose.call(this);
    }
});
