// syracuse-soap/lib/server/representations/soapWebService.js
"use strict";
exports.representation = {
	resources: function() {
		var locale = require("syracuse-core/lib/locale");
		var globals = require('streamline-runtime').globals;
		return locale.resources(module, globals.context && globals.context.sessionLocale)();
	},
	$entityName: "soapWebService",
	$facets: {
		$details: {
			$layout: {
				$items: [{
					$title: "{@InformationsSectionTitle}",
					"$category": "section",
					$layout: {
						$items: [{
							$bind: "name"
						}, {
							$bind: "description"
						}, {
							$bind: "wsdl"
						}, {
							$bind: "url"
						}]
					}
				}, {
					$title: "{@AdministrationSectionTitle}",
					"$category": "section",
					$layout: {
						$items: [{
							$bind: "application"
						}, {
							$bind: "endpoint"
						}]
					}
				}, {
					$title: "{@ContentSectionTitle}",
					"$category": "section",
					$layout: {
						$items: [{
							$bind: "representations"
						}]
					}
				}, {
					$title: "{@TesterSectionTitle}",
					"$category": "section",
					$layout: {
						$items: [{
							$bind: "stub",
							$layout: {
								$items: [{
									$bind: "ports",
									$layout: {
										$items: [{
											$bind: "operations"
										}]
									},
									$opened: true
								}]
							}
						}]
					}
				}]
			}
		},
		$edit: {
			$layout: {
				$items: [{
					$title: "{@InformationsSectionTitle}",
					"$category": "section",
					$layout: {
						$items: [{
							$bind: "name"
						}, {
							$bind: "description"
						}]
					}
				}, {
					$title: "{@AdministrationSectionTitle}",
					"$category": "section",
					$layout: {
						$items: [{
							$bind: "application"
						}, {
							$bind: "endpoint"
						}]
					}
				}, {
					$title: "{@ContentSectionTitle}",
					"$category": "section",
					$layout: {
						$items: [{
							$bind: "representations"
						}]
					}
				}]
			}
		},
		$query: {
			$layout: {
				$items: [{
					$category: "section",
					$layout: {
						$items: [{
							$category: "section",
							$layout: {
								$items: [{
									$bind: "$resources",

								}]
							}
						}]
					}
				}]
			}
		}
	}
};
