// syracuse-ui/lib/tests/pages/notifications/testNotif.js
"use strict";
var SocketHandler = require('syracuse-ui/lib/notifications/socketHandler');


exports.load = function(page, finish) {
	var socketHandler = SocketHandler.create();
	var col = page.addCol(page.layoutContent.layoutSlot);
	var slotConnected = col.appendChild(document.createElement("div"));
	var slotevent = col.appendChild(document.createElement("div"));
	var dynamicEvent = col.appendChild(document.createElement("div"));
	var socketDynamic;

	var socket = socketHandler.register("/unit_test", {
		eventClient1: function(data) {
			if (data === 'data') {
				slotevent.innerHTML = "string 'data' received";

			} else {
				slotevent.innerHTML = "string 'data' not received";

			}
		},
		eventClient2: function(data) {
			if (data && data.ok) {
				slotevent.innerHTML = "object '{ok : true}' received";
			} else {
				slotevent.innerHTML = "object '{ok : true}' not received";

			}

		},
		delDynEvent: function(data) {
			// remove button
			syra_button.hide(dina1, true);
			syra_button.hide(dina2, true);
			if (socketDynamic) {
				socketHandler.unregister(data);
				socketDynamic = null;
			}
		},
		connectTest: function(data) {
			slotConnected.innerHTML = "socket connected to unit test event";
		}
	});

	page.onTestButtonClick2 = function() {
		socket.emit("event1", 'data');
	};
	page.onTestButtonClick3 = function() {
		socket.emit("event2", {
			ok: true
		});
	};
	page.onTestButtonClick4 = function() {
		socket.emit("unregisterEvent");
	};
	var dina1;
	var dina2;

	page.dynamic1 = function() {
		socketDynamic.emit("dynamicEvent1", 'data');
	};
	page.dynamic2 = function() {
		socketDynamic.emit("dynamicEvent2", {
			ok: true
		});
	};



	var dynamicUnit;
	page.onTestButtonClick1 = function() {
		// register event to be able to receive them
		dynamicUnit = "/unit_test" + (new Date()).getTime();

		socket.emit("registerEvent", dynamicUnit);
		socketDynamic = socketHandler.register(dynamicUnit, {
			eventDynamicClient1: function(data) {
				if (data === 'data') {
					slotevent.innerHTML = "dynamic event : string 'data' received";

				} else {
					slotevent.innerHTML = "dynamic event : string 'data' not received";
				}
			},
			eventDynamicClient2: function(data) {
				if (data && data.ok) {
					slotevent.innerHTML = "dynamic event : object '{ok : true}' received";
				} else {
					slotevent.innerHTML = "dynamic event : object '{ok : true}' not received";
				}
			},
			connectTest: function(data) {
				slotConnected.innerHTML = "socket connected to dynamic unit test event";
				dina1 = dina1 || syra_button.add({
					parent: page,
					slot: page.layoutContent.domItem,
					text: "dynamicEvent1",
					css: "s-uitest-mn",
					click: page.dynamic1,
					isHidden: true
				});
				dina2 = dina2 || syra_button.add({
					parent: page,
					slot: page.layoutContent.domItem,
					text: "dynamicEvent2",
					css: "s-uitest-mn",
					click: page.dynamic2,
					isHidden: true
				});
				syra_button.hide(dina1, false);
				syra_button.hide(dina2, false);
			}
		});
	};
	syra_button.add({
		parent: page,
		css: "s-uitest-mn",
		text: "testEvent1",
		slot: page.layoutContent.domItem,
		click: page.onTestButtonClick2
	});
	syra_button.add({
		parent: page,
		css: "s-uitest-mn",
		slot: page.layoutContent.domItem,
		text: "testEvent2",
		click: page.onTestButtonClick3
	});
	syra_button.add({
		parent: page,
		slot: page.layoutContent.domItem,
		css: "s-uitest-mn",
		text: "addDynamicEvent",
		click: page.onTestButtonClick1
	});
	syra_button.add({
		parent: page,
		slot: page.layoutContent.domItem,
		css: "s-uitest-mn",
		text: "removeDynamicEvent",
		click: page.onTestButtonClick4
	});

	finish(page);

};
