// syracuse-ui/lib/authoring/list/cardDesigner.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var DesktopPageDesigner = require("syracuse-ui/lib/authoring/desktopPage/desktopPageDesigner").DesktopPageDesigner;

function CardDesigner() {}

exports.CardDesigner = helpers.defineClass(CardDesigner, DesktopPageDesigner, {
	initializePageDesigner: function() {
		this.togglePageBars(true);
		this.awArticle.$designLevel = "article";
		DesktopPageDesigner.prototype.initializePageDesigner.call(this);
	},
	openWorkingCopy: function() {
		var link = document.createElement("a");
		link.setAttribute("data-s-picker", link.syraPickerType = "close-designer");
		link.title = this.localize.aw_close;
		link.className = "s-aw-carddesigner-close";
		this.workinCopySlot.appendChild(link);
	},
	drawBox: function() {
		DesktopPageDesigner.prototype.drawBox.call(this);
		document.site.removeDomChild(this.viewsSlot);
		this.viewsSlot = null;
	},
	onClickPicker: function(picker) {
		if (picker.syraPickerType == "close-designer") {
			var self = this;
			setTimeout(function() {
				self.closeDesigner();
			}, 100);
			return false;
		}
		return DesktopPageDesigner.prototype.onClickPicker.call(this, picker);
	},
	closeDesigner: function() {
		if (this.historyTool._currentStep > 1) {
			this.awArticle.list.designer.historyTool.notifyUpdate(true);
		}
		document.site.designItem(false, this.awArticle);
	},
	togglePageBars: function(designing) {
		var page = this.awArticle.page;
		if (!page.designer) {
			if (page.menuBar) {
				page.menuBar.toggleUIDesign(designing);
			}
			if (page.fusionBar) {
				page.fusionBar.toggleUIDesign(designing);
			}
		}
	},
	endTargetPageUpdate: function(selectedItem, isStructureUpdated) {
		this.awArticle.page.layoutValidator.validate(this.awArticle.layoutContent, true);
		this.toggleUIDesign(this.awArticle, true);
		DesktopPageDesigner.prototype.endTargetPageUpdate.call(this, selectedItem, isStructureUpdated);
	},
	dispose: function() {
		this.togglePageBars(false);
		DesktopPageDesigner.prototype.dispose.call(this);
	}
});
