// syracuse-ui/lib/authoring/list/cardDesigner.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var DesktopPageDesigner = require("syracuse-ui/lib/authoring/desktopPage/desktopPageDesigner").DesktopPageDesigner;

function CardDesigner() {}

exports.CardDesigner = helpers.defineClass(CardDesigner, DesktopPageDesigner, {
	initializePageDesigner: function() {
		this.togglePageBars(true);
		this.awArticle.$designLevel = "article";
		DesktopPageDesigner.prototype.initializePageDesigner.call(this);
	},
	togglePageBars: function(designing) {
		var page = this.awArticle.page;
		if (!page.designer) {
			if (page.menuBar) {
				page.menuBar.toggleUIDesign(designing);
			}
			if (page.fusionBar) {
				page.fusionBar.toggleUIDesign(designing);
			}
		}
	},
	endTargetPageUpdate: function(selectedItem, isStructureUpdated) {
		this.awArticle.page.layoutValidator.validate(this.awArticle.layoutContent, true);
		this.toggleUIDesign(this.awArticle, true);
		DesktopPageDesigner.prototype.endTargetPageUpdate.call(this, selectedItem, isStructureUpdated);
	},
	/*loadDefaultTools: function(css){
     this.historyTool = this.awArticle.list.designer.historyTool;
     this.designerWorkingCopy = this.awArticle.list.designer.designerWorkingCopy;
     },*/
	dispose: function() {
		this.togglePageBars(false);
		this.historyTool = this.designerWorkingCopy = null;
		DesktopPageDesigner.prototype.dispose.call(this);
	}
});
