// syracuse-ui/lib/authoring/list/cardDesigner.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var DesktopPageDesigner = require("syracuse-ui/lib/authoring/desktopPage/desktopPageDesigner").DesktopPageDesigner;

function CardDesigner() {}

exports.CardDesigner = helpers.defineClass(CardDesigner, DesktopPageDesigner, {
	initializePageDesigner: function() {
		this.togglePageBars(true);
		this.awArticle.$designLevel = "article";
		DesktopPageDesigner.prototype.initializePageDesigner.call(this);
	},
	togglePageBars: function(designing) {
		var page = this.awArticle.page;
		if (!page.designer) {
			if (page.menuBar) {
				page.menuBar.toggleUIDesign(designing);
			}
			if (page.fusionBar) {
				page.fusionBar.toggleUIDesign(designing);
			}
		}
	},
	getAwItemsqs: function(awItem) {
		return awItem == this.awArticle ? this.awArticle.cardItem : awItem;
	},
	selectItemsqs: function(item, select, ensureSelect) {
		if (item == this.awArticle.cardItem) {
			item = this.awArticle;
		}
		DesktopPageDesigner.prototype.selectItem.call(this, item, select, ensureSelect);
	},
	sqsqtoggleUIDesign: function(item, enable) {
		DesktopPageDesigner.prototype.toggleUIDesign.call(this, this.getAwItem(item), enable);
	},
	toggleUIDesign: function(item, enable) {
		DesktopPageDesigner.prototype.toggleUIDesign.call(this, item, enable);
	},
	endTargetPageUpdate: function(selectedItem, isStructureUpdated) {
		if (selectedItem && selectedItem.disposed) {
			selectedItem = null;
		}
		this.toggleUIDesign(this.awArticle, true);
		DesktopPageDesigner.prototype.endTargetPageUpdate.call(this, selectedItem, isStructureUpdated);
	},
	endTargetPageUpdatesqsq: function(selectedItem, isStructureUpdated) {
		this.hookCardItem();
		if (selectedItem && selectedItem.disposed) {
			selectedItem = null;
		}
		this.toggleUIDesign(this.awArticle, true);
		DesktopPageDesigner.prototype.endTargetPageUpdate.call(this, selectedItem, isStructureUpdated);
	},
	/*loadDefaultTools: function(css){
     this.historyTool = this.awArticle.list.designer.historyTool;
     this.designerWorkingCopy = this.awArticle.list.designer.designerWorkingCopy;
     },*/
	dispose: function() {
		this.togglePageBars(false);
		this.historyTool = this.designerWorkingCopy = null;
		DesktopPageDesigner.prototype.dispose.call(this);
	}
});
