// syracuse-core/lib/types/integer.js
"use strict";
var locale = require("syracuse-core/lib/locale");
var decimal = require("syracuse-core/lib/types/decimal");

var numberFormat = require("./numberFormat");

exports.validate = function(value, constraints, errors) {
	if (value == null) {
		if (constraints.$isMandatory) return errors.push(locale.format(module, "valMan"));
		if (!constraints.$isNullable) return errors.push(locale.format(module, "valNull"));
		return;
	}
	if (typeof value == "string") {
		value = parseInt(value);
	} else if (typeof value !== "number" || (value % 1 !== 0)) {
		errors.push(locale.format(module, "valInt"));
		return;
	}
	if (constraints.$divisibleBy != null)
		if (value % constraints.$divisibleBy !== 0) errors.push(locale.format(module, "valDiv", constraints.$divisibleBy));
	if (constraints.$enum != null && !constraints.$enum.some(function(elt) {
		return elt.$value == value;
	})) errors.push(locale.format(module, "valEnum"));
	if (constraints.$minimum != null) {
		if (constraints.$minimumCanEqual) {
			if (value < constraints.$minimum) errors.push(locale.format(module, "valGe", constraints.$minimum));
		} else {
			if (value <= constraints.$minimum) errors.push(locale.format(module, "valGt", constraints.$minimum));
		}
	}
	if (constraints.$maximum != null) {
		if (constraints.$maximumCanEqual) {
			if (value > constraints.$maximum) errors.push(locale.format(module, "valLe", constraints.$maximum));
		} else {
			if (value >= constraints.$maximum) errors.push(locale.format(module, "valLt", constraints.$maximum));
		}
	}
};

exports.parse = function(str, format) {
	return Math.floor(numberFormat.parse(str, format));
};

exports.format = function(val, format) {
	return numberFormat.format(val, format);
};

exports.add = function(n1, n2) {
	return n1 + n2;
};

exports.subtract = function(n1, n2) {
	return n1 - n2;
};

exports.numberValue = function(val) {
	return parseInt(val, 10);
};

exports.defaultValue = 0;

// ==== old parse method ===

//exports.parse = function(value, gs, ds){
//    gs = (gs === ".") ? "[.]" : gs;
//    return decimal.make(value.toString().replace(new RegExp(gs || " ", "g"), ""), 0);
//}
