// syracuse-core/lib/types/real.js
"use strict";
var locale = require("syracuse-core/lib/locale");
var numberFormat = require("./numberFormat");
exports.validate = function(value, constraints, errors) {
	if (value == null) {
		if (constraints.$isMandatory) return errors.push(locale.format(module, "valMan"));
		if (!constraints.$isNullable) return errors.push(locale.format(module, "valNull"));
		return;
	}
	if (typeof value !== "number" || isNaN(value)) {
		errors.push(locale.format(module, "valNumber"));
		return;
	}
	if (constraints.$divisibleBy != null)
		if (value % constraints.$divisibleBy !== 0) errors.push("value must be divisible by" + constraints.$divisibleBy);
	if (constraints.$enum != null && !constraints.$enum.some(function(elt) {
		return elt.$value == value;
	})) errors.push(locale.format(module, "valEnum"));
	if (constraints.$minimum != null) {
		if (constraints.$minimumCanEqual) {
			if (value < constraints.$minimum) errors.push(locale.format(module, "valGe", constraints.$minimum));
		} else {
			if (value <= constraints.$minimum) errors.push(locale.format(module, "valGt", constraints.$minimum));
		}
	}
	if (constraints.$maximum != null) {
		if (constraints.$maximumCanEqual) {
			if (value > constraints.$maximum) errors.push(locale.format(module, "valLe", constraints.$maximum));
		} else {
			if (value >= constraints.$maximum) errors.push(locale.format(module, "valLt", constraints.$maximum));
		}
	}
};

exports.defaultValue = 0;

exports.format = function(val, format) {
	return numberFormat.format(val, format);
};

exports.parse = function(str, format) {
	return numberFormat.parse(str, format);
};

exports.add = function(n1, n2) {
	return n1 + n2;
};

exports.subtract = function(n1, n2) {
	return n1 - n2;
};

exports.numberValue = function(val) {
	return Number(val);
};
