// @sage/syracuse-react/src/profile/bar/select/roleChoice.tsx
"use strict";
import * as React from 'react';
import { PopupLink } from '../../../over/popup/popupLink'
import { Link } from '../../../link/link';
import { ScrollList } from '../../../common/list/list';
import { HotKeyLinks } from '../../../shortcut/hotKeyLinks';
import * as SyraPageModal from '../../../over/modal/page/syraPageModal';
import { showDiagnoses } from '../../../alert/alert';
import { getLocalString } from '../../../culture/localString';
import { Profile, IRole } from '../../profile';

export class RoleChoice extends React.Component<{
    profile: Profile,
    onTogglePopup?: (opened: boolean) => void,
    onSelect: (record: any) => void
}, {}> {
    popupLink: PopupLink;
    list: ScrollList;
    onPopupResize = (maxHeight: number) => {
        if (this.list) {
            let node = this.list.rootNode.previousSibling as HTMLElement;
            if (node) {
                maxHeight -= node.clientHeight;
            }
            node = this.list.rootNode.nextSibling as HTMLElement;
            if (node) {
                maxHeight -= node.clientHeight;
            }
            this.list.listHeight = maxHeight;
        }
    }
    onMoreClick = () => {
        this.popupLink && this.popupLink.togglePopup();
        SyraPageModal.openSelectModal({
            authoringDisabled: true,
            url: this.props.profile.roleLookupUrl,
            onLoadRepresentation: (diagnoses) => diagnoses && showDiagnoses(diagnoses),
            onValidateSelection: (selected: any[]) => this.props.onSelect(selected[0])
        });
    }
    renderMore() {
        return <div className="s_profile_bar_list_item_primary_bottom">
            <Link
                data-s-profile-select-more="1"
                className="s_profile_bar_list_item s_primary"
                title={getLocalString("profile_role_details")}
                onClick={this.onMoreClick} />
        </div>
    }
    componentWillUnmount() {
        this.list = this.popupLink = null;
    }
    onSelectedClick = () => {
        this.popupLink && this.popupLink.togglePopup();
    }
    renderPopup = () => {
        let selectedRole = this.props.profile.selectedRole;
        let records: any[] = [];
        if (this.props.profile.roles) {
            let selectedUuid = this.props.profile.selectedRoleUuid;
            for (let role of this.props.profile.roles) {
                if (selectedUuid != role.$uuid) {
                    records.push(<Link
                        key={role.$uuid}
                        className="s_profile_bar_list_item"
                        title={role.description}
                        value={role}
                        onClick={this.props.onSelect} />);
                }
            }
        }
        return <HotKeyLinks>
            {selectedRole &&
                <Link
                    className="s_profile_bar_list_item s_selected"
                    value={selectedRole}
                    onClick={this.onSelectedClick}>
                    {selectedRole.description}
                </Link>}
            <ScrollList className="s_profile_bar_list" ref={(node) => this.list = node}>
                {records}
            </ScrollList>
            {this.props.profile.roleLookupUrl && this.renderMore()}
        </HotKeyLinks>;
    }
    componentDidUpdate(prevProps: any) {
        if (this.popupLink && this.popupLink.popup) {
            this.popupLink.popup.onResize();
        }
    }
    render() {
        let profile = this.props.profile;
        return <PopupLink
            ref={(node) => { this.popupLink = node }}
            data-s-profile-tab="selectedRole"
            className="s_profile_bar_textlink"
            title={profile.selectedRole && profile.selectedRole.description}
            onTogglePopup={this.props.onTogglePopup}
            popup={{
                onResize: this.onPopupResize,
                render: this.renderPopup
            }} />;
    }
}
