// @sage/syracuse-react/src/profile/bar/select/roleSelect.tsx
"use strict";
import * as React from 'react';
import { Dropdown } from './dropdown';
import { IRole } from '../../profile';
import { getLocalString } from '../../../culture/localString';
import { IProfileRoleSelect } from '../profileBarProps';

interface IRoleSelectProps extends IProfileRoleSelect {
    lookupUrl: string,
    selected: IRole,
    roles: IRole[]
}

export class RoleSelect extends React.Component<IRoleSelectProps, {
}> {
    getRecords = () => {
        let unSelected: any[] = [];
        let selectedId = this.props.selected && this.props.selected.$uuid;
        for (let role of this.props.roles) {
            if (selectedId != role.$uuid) {
                unSelected.push(role);
            }
        }
        return {
            title: "description",
            selected: this.props.selected,
            unselected: unSelected
        };
    }
    renderItem = (role: IRole) => {
        return role && role.description;
    }
    render() {
        return <Dropdown
            qualityBind="selectedRole"
            title={this.props.selected && this.props.selected.description}
            className="s_profile_bar_textlink"
            getRecords={this.getRecords}
            renderItem={this.renderItem}
            moreUrl={this.props.lookupUrl}
            onSelected={this.props.onSelectRole}
            moreTitle={getLocalString("profile_role_details")} />;
    }
}
