// @sage/syracuse-ui/lib/field/reference/lookup.js
"use strict";
var List = require('@sage/syracuse-ui/lib/scroll/list').List;

function Lookup(field) {
	this.field = field;
	this.id = field.id + "-lookup";
	syra_item.register(this);
	if ((field.articleParent && field.articleParent.onLookupClick) ? field.articleParent.onLookupClick(field) : true) {
		field.focus && field.focus();
		this.fetch();
	}
}

Lookup.prototype.fetch = function() {
	var self = this;
	syra_over.openSelectModal({
		authoringDisabled: true,
		url: syra_expression.parse(self.field.articleParent, self.field.$menus.$lookup.$url),
		onLoadRepresentation: function(diagnoses) {
			diagnoses && syra_alert.show(diagnoses);
		},
		onValidateSelection: function(selected) {
			self.select(selected[0]);
		}
	});
};
Lookup.prototype.click = function(event, target) {
	var self = this;
	if (target.syraManage) {
		self.page.startChange();
		self.page.applyChange(self.data);
		self.page.endChange();
		self.page.menuBar && self.page.menuBar.toggleBar(false);
		syra_over.renderPageModal(self.page, {
			authoringDisabled: true,
			onSelectRecord: function(records) {
				self.select(records[Object.keys(records)[0]].dataset);
			},
			close: function() {
				setTimeout(function() {
					self.dispose();
				}, 100);
			}
		});
	} else {
		if (target.syraRecord !== undefined) {
			self.select(self.data[self.$arrayBind][target.syraRecord]);
			self.dispose();
		}
	}

};

Lookup.prototype.select = function(dataset) {
	if (!this.field.$isDisabled && !this.field.$isReadOnly) {
		if (this.field.articleParent && this.field.$reference) {
			var parentDataset = this.field.articleParent.dataset;
			if (parentDataset && parentDataset.$properties) {
				var $prop = parentDataset.$properties[this.field.$reference.$value.$prop];
				if ($prop) {
					delete $prop.$diagnoses;
				}
			}
		}
		if (this.field.isChildFieldRecord) {
			this.field.articleParent.setValue(dataset, {
				$diagnoses: null
			});
			syra_form.updateDelta(this.field.articleParent, dataset);
		} else {
			this.field.setValue((dataset && this.field.$menus.$lookup.$result) ? dataset[this.field.$menus.$lookup.$result] : dataset, {
				$diagnoses: null
			});
			this.field.focus();
			syra_form.update(this.field, this.field.currentValue);
		}
	}
};
Lookup.prototype.dispose = function() {
	if (this.field) {
		delete this.field._lookupItem;
	}
	this._popup && this._popup.close();
	this.page && this.page.dispose();
	syra_item.unregister(this);
	syra_article.dispose(this);
};

function _clickPicker(options, event) {
	var field = this.parent;
	if (field._lookupItem) {
		field._lookupItem.dispose();
	} else {
		field._lookupItem = new Lookup(field);
	}
}

exports.addPicker = function(field, $menu) {
	if (field.$isEditMode && !field.picker_$lookup) {
		syra_button.addFieldPicker({
			parent: field,
			pickerId: "$lookup",
			fontIcon: "lookup",
			text: syra_local.fieldLookup,
			shortCutTip: syra_shortcut.tip.openLookup,
			isDisabled: $menu.$isDisabled,
			click: _clickPicker
		});
	}
	$menu.$isPopup = false;
};
