// syracuse-soap/lib/client/entities/soapStubOperation.js
"use strict";

exports.entity = {
	$properties: {
		name: {
			$title: "Name",
			$isMandatory: true,
			$isUnique: true
		},
		description: {
			$title: "Description",
		},
		tester: {
			$title: "Tester",
			$format: "$url",
			$isHidden: true
		},
		soapAction: {
			$title: "Soap Action",
		},
		style: {
			$title: "Style",
		},
		location: {
			$title: "Location",
		},
		inputHeaderEncoding: {
			$title: "Input header encoding",
		},
		inputHeaderName: {
			$title: "Input header name",
		},
		inputBodyEncoding: {
			$title: "Input body encoding",
		},
		inputBodyName: {
			$title: "Input body name",
		},
		requestPrototype: {
			$title: "Request prototype",
			$isMandatory: true,
			$type: "binary",
			$storage: "db_file",
		},
		outputHeaderEncoding: {
			$title: "Output header encoding",
		},
		outputHeaderName: {
			$title: "Output header name",
		},
		outputBodyEncoding: {
			$title: "Output body encoding",
		},
		outputBodyName: {
			$title: "Output body name",
		},
		responsePrototype: {
			$title: "Response prototype",
			$isMandatory: true,
			$type: "binary",
			$storage: "db_file",
		},
	},
	$relations: {},
	$defaultOrder: [
		["name", true]
	]
};
