// syracuse-ui/lib/events/shortCutInvoker.js
"use strict";
var keyCodeMap = require('syracuse-ui/lib/events/keyboardHandlers').keyCodeMap;

var menuStack = [];

function _clearMenuStack(item) {
	if (item._lastItem) {
		_walkMenuItem(item._lastItem, false);
	}
	menuStack = [];
}

function _listChilds(menus) {
	var menuItem;

	if (menus && !menus.isMenuGroupRoot) {
		menuStack.push(menus);
	}

	if (menus.boxChildItems) {
		for (var mm = 0, kk = menus.boxChildItems.length; mm < kk; mm++) {
			menuItem = menus.boxChildItems[mm];
			if (!menuItem.$isDisabled && !menuItem.$isHidden && (menus.$opened || menus.isMenuItem)) {
				_listChilds(menuItem);
			}
		}
	}
}

function _walkMenuItem(menuItem, show) {
	var classGroup = menuItem && menuItem.domTitle || menuItem && menuItem.domItem;

	if (menuItem && menuItem.domItem) {
		syra_site.dom.toggleClass(classGroup, "s-enter", show);
	}
	return;
}

function _walkActionPanel(item, shortcuts, event) {
	menuStack = [];

	item = item.page;

	if (item.menuBarLayout && item.menuBarLayout.items) {
		for (var ii = 0, jj = item.menuBarLayout.items.length; ii < jj; ii++) {
			var menus = item.menuBarLayout.items[ii];
			if (menus.boxChildItems && (menus.$opened || menus.isMenuItem)) {
				_listChilds(menus);
			}
		}
	}

	if (menuStack.length >= 0) {
		if (!item._lastItem) {
			item._lastItem = menuStack[0];
			return _walkMenuItem(item._lastItem, true);
		}

		// remove old marking
		_walkMenuItem(item._lastItem, false);


		// new marking
		for (var kk = 0, mm = menuStack.length - 1; kk <= mm; kk++) {
			if (menuStack[kk] === item._lastItem) {
				if (shortcuts.down) {
					if (kk < mm) {
						item._lastItem = menuStack[kk + 1];
					} else {
						item._lastItem = menuStack[0];
					}
				}

				if (shortcuts.up) {
					if (kk > 0) {
						item._lastItem = menuStack[kk - 1];
					} else {
						item._lastItem = menuStack[mm];
					}
				}
				if (shortcuts.left) {
					if (item._lastItem.$opened === true) {
						item._lastItem.openBox(false);
					}
				}
				if (shortcuts.right) {
					if (item._lastItem.$opened === false) {
						item._lastItem.openBox(true);
					}
				}
				return _walkMenuItem(item._lastItem, true);
			}
		}
	}

	return true;
}

function _walkTabs(tab, shortcuts, focusedTab) {
	if (focusedTab && focusedTab.layoutParent == tab.layoutParent) {
		tab = focusedTab;
	}
	var items = tab.layoutParent.items;
	var max = items.length - 1;
	var ii = items.indexOf(tab);
	while (true) {
		if (shortcuts.right) {
			if (ii == max) {
				ii = 0;
			} else {
				ii++;
			}
		} else {
			if (ii == 0) {
				ii = max;
			} else {
				ii--;
			}
		}
		if (!items[ii].isInvisible) {
			tab.page.focusedTab = items[ii];
			items[ii].header.click();
			return;
		}
	}
}

function _walkPageTabs(item, shortcuts, event) {
	var section;
	var page = item.page;
	if (page == syra_site.landingPageMaster) {
		page = page.getSelectedChildPage() || page;
	}
	if (event.shiftKey || (page == item)) { //ESC + SHIFT + LEFT-RIGHT, or page level 
		for (var ii = 0, jj = page.childrenSection.length - 1; ii < jj; ii++) {
			section = page.childrenSection[ii];
			if (section.isTabLayout) {
				_walkTabs(section, shortcuts, page.focusedTab);
				return;
			}
		}
	}
	//tab not found at top level, walk from item level
	section = item.boxParent;
	while (section) {
		if (section.isTabLayout) {
			_walkTabs(section, shortcuts, page.focusedTab);
			return;
		}
		section = section.boxParent;
	}
	if (page.focusedTab) {
		_walkTabs(page.focusedTab, shortcuts);
	} else {
		(page != item) && _walkPageTabs(page, shortcuts, event); //test to avoid loop on page
	}
}

function _resizeGridColumns(item) {
	var hasFit;
	var lists = (item.page == item) ? syra_store.getLists(item) : [];
	item.isArrayField && lists.push(item);
	if (lists.length) {
		for (var ii = 0, jj = lists.length; ii < jj; ii++) {
			var builder = lists[ii] && lists[ii].builder;
			if (builder && builder.toggleColWidthToContent) {
				hasFit = true;
				builder.toggleColWidthToContent();
			}
		}
	}
	return hasFit;
}

function _selectListHTMLContent(list) {
	var range = document.createRange();
	var scrollHeadTable = list.builder.scrollTable && list.builder.scrollTable.headTable;
	scrollHeadTable && range.selectNode(scrollHeadTable);
	list.body && range.selectNode(list.body);
	var select = document.getSelection();
	select.removeAllRanges();
	select.addRange(range);
}

function _togglePageSideBar(page, switchLeft, switchRight) {
	for (var ii = 0, jj = page.bars.length; ii < jj; ii++) {
		var bar = page.bars[ii];
		if ((switchLeft && bar.options.resizeDirection == "left") ||
			(switchRight && bar.options.resizeDirection != "left")) {
			bar.collapse();
		}
	}
}

function _resetUserPreferences(page) {
	if (page == syra_site.landingPageMaster) {
		page = page.getSelectedChildPage() || page;
	}
	if (page) {
		syra_diagnose.box.show({
			$title: syra_local.aw_deleteMessageTitle,
			$message: syra_local.aw_deletePreferences,
			$type: "question",
			$buttons: "yesno",
			callback: function(response) {
				if (response && response.$clientId == "yes") {
					page.userPreferences && page.userPreferences.reset();
					if (page.isLandingPage) {
						var $binds = Object.keys(page.boundFields);
						for (var ii = 0, jj = $binds.length; ii < jj; ii++) {
							var vignette = page.boundFields[$binds[ii]];
							vignette = vignette && vignette[0] && vignette[0].vignette;
							vignette && vignette.userPreferences && vignette.userPreferences.reset();
						}
					}
					for (var ii = 0, jj = page.bars.length; ii < jj; ii++) {
						page.bars[ii].clearPreferences();
					}
					var lists = syra_store.getLists(page);
					for (var ii = 0, jj = lists.length; ii < jj; ii++) {
						lists[ii].clearPreferences && lists[ii].clearPreferences();
					}
					page.resizeArticle(true);
				}
			}
		});
	}
}

var _dynRegisteredSC = [];

function _invokeShortCut(shortcutCbck) {
	var ret;
	if (shortcutCbck.bind) {
		ret = shortcutCbck.callback.call(shortcutCbck.bind, event, shortcutCbck.data);
	} else {
		ret = shortcutCbck.callback(event, shortcutCbck.data);
	}
	return ret === false || ret === null ? false : true;
}

function _isRegisteredShortCut(shortcutFired, item) {
	var i, j, leni, lenj, sc, gotIt = false;
	var scKeys = Object.keys(shortcutFired),
		exludedSCProps = {
			"isEnabled": true
		};
	var tot1 = 0,
		tot2, tot3;
	// Check the incoming shortcut object
	for (i = 0, leni = scKeys.length; i < leni; i++) {
		if (!exludedSCProps[scKeys[i]] && shortcutFired[scKeys[i]]) {
			tot1++;
		}
	}
	// Is the incoming shortcut has been registered for the target item?
	for (i = 0, leni = _dynRegisteredSC.length; i < leni; i++) {
		if ((sc = _dynRegisteredSC[i].scDef) && sc.item === item) {
			scKeys = Object.keys(sc);
			tot2 = scKeys.length;
			if (sc.hasOwnProperty("item")) {
				tot2--;
			}
			tot3 = 0;
			for (j = 0, lenj = scKeys.length; j < lenj; j++) {
				if (scKeys[j] != "item") {
					if (!shortcutFired[scKeys[j]]) {
						break;
					} else
						tot3++;
				}
			}
			if (tot1 == tot2 && tot1 == tot3) {
				gotIt = true;
				break;
			}
		}
	}
	return gotIt ? _dynRegisteredSC[i] : null;
}
exports.registerShortCut = function(item, scDef, id, callback, callbackBind, callbackData) {
	if (!scDef.item) {
		scDef.item = item;
	} else if (scDef.item === "page") {
		scDef.item = item.page;
	}
	_dynRegisteredSC.push({
		"scDef": scDef,
		"cbck": {
			"callback": callback,
			"bind": callbackBind,
			"data": callbackData
		},
		"key": id
	});
};
exports.unregisterShortCut = function(item, id) {
	var i, len, sc;
	for (i = 0, len = _dynRegisteredSC.length; i < len; i++) {
		sc = _dynRegisteredSC[i];
		if (sc.key === id && sc.scDef && sc.scDef.item === item) {
			sc.cbck = null;
			sc.scDef.item = null;
			sc.scDef = null;
			_dynRegisteredSC.splice(i, 1);
			break;
		}
	}
};

exports.applyEscape = function(item, shortcuts, event) {
	var isApplyed;
	if (item.applyEscape && item.applyEscape(shortcuts, event)) { //close choice, date, time selectors
		return true;
	}
	if (item.isField) {
		if (item.popupPicker) {
			item.popupPicker.close();
			isApplyed = true;
		} else {
			if (item.page.externalAdapter && item.page.externalAdapter.applyEscape) {
				isApplyed = item.page.externalAdapter.applyEscape(item);
			}
		}
	}!isApplyed && item.page.dialogWrapper && item.page.dialogWrapper.close();
};

exports.apply = function(item, shortcuts, event) {
	if (item.variantItem) {
		return exports.apply(item.variantItem, shortcuts, event);
	}
	var isPage = item.page == item,
		dynSC;
	if (shortcuts.esc) {
		if (isPage) {
			// Just check if it's a dynamic registered shortcut (only for page today...)
			if ((dynSC = _isRegisteredShortCut(shortcuts, item)) && _invokeShortCut(dynSC.cbck)) {
				return true;
			}
		}
		if (shortcuts.f1) {
			//open help on item, SHIFT open help on page
			syra_site.helpCenter.openByShortCut(item.page, (event.shiftKey || isPage) ? null : item);
			return true;
		}
		if (shortcuts.f5) {
			//refresh page data with autofetch mode
			if (isPage) {
				syra_menus.click.refreshPageData(item, event);
				return true;
			}
		}
		if (shortcuts.f6) {
			//showProperty of field. Currently for fusion. Syracuse to do
			syra_menus.click.showFieldProperty(item);
			return true;
		}
		if (shortcuts.f9) {
			//open fusion tunnel
			return item.isField && syra_menus.click.tunnel(item);
		}
		if (shortcuts.f11 || (shortcuts.h && !shortcuts.g)) {
			//switch left /right side bar
			var switchLeft = shortcuts.f11 || (shortcuts.h && (shortcuts.a || shortcuts.l));
			var switchRight = shortcuts.f11 || (shortcuts.h && (shortcuts.a || shortcuts.r));
			if (switchLeft || switchRight) {
				_togglePageSideBar(item.page, switchLeft, switchRight);
			}
			return true;
		}
		if (shortcuts.q) {
			// put cursor into first editable field of filter row
			var list;
			if (isPage && !item.isFusionPage) {
				list = syra_store.findList(item.page);

			} else {
				list = item;
			}
			if (list) {
				var filterRecord = list.builder && list.builder.filterRecord;
				filterRecord && filterRecord.setFocus();
			}
			return true;
		}
		if (shortcuts.m || shortcuts.f4) {
			//open contextual menu
			return syra_menus.click.contextualMenu(event.shiftKey ? item.articleParent : item);
		}
		if (shortcuts.insert || (shortcuts.n && !shortcuts.g)) {
			//trigger create action
			var page = item;
			if (page == syra_site.landingPageMaster) {
				page = page.getSelectedChildPage() || page;
			}
			if (page.isLandingPage) {
				if (event.shiftKey) {
					page.addVignetteBtn.link.click();
				} else {
					page.addTabBtn.link.click();
				}
				return true;
			}
			return syra_menus.click.create(item);
		}
		if (shortcuts["delete"]) {
			return syra_menus.click.deleteItem(item);
		}
		if (shortcuts.enter) {
			//trigger main action of page
			var main = isPage && item.getMainAction();
			main && syra_menus.click.menuItem(main, event);
			return true;
		}
		if (shortcuts.a) {
			if (isPage && event.ctrlKey) {
				item.selectText && item.selectText();
				return true;
			}
			if (item.isArrayField && item.selectHTMLContent) {
				_selectListHTMLContent(item);
				return true;
			}
		}
		if (isPage) {
			if (shortcuts.p) {
				syra_menus.click.print(item);
				return true;
			}
			if (shortcuts.j) {
				shortcuts.j && syra_menus.click.prev(item, event.shiftKey);
				return true;
			}
			if (shortcuts.k) {
				syra_menus.click.next(item, event.shiftKey);
				return true;
			}
		}
		if (shortcuts.w) {
			shortcuts.z && _resetUserPreferences(item.page);
			return true;
		}
		if (shortcuts.g) {
			shortcuts.h && syra_site.homeLink.click();
			shortcuts.n && syra_menus.click.navigation();
			return true;
		}
		if (item.isField && ((!shortcuts.h && shortcuts.l) || shortcuts.f12)) {
			syra_menus.click.selector(item);
			return true;
		}
		if (shortcuts.r) {
			return _resizeGridColumns(item);
		}

		if (item.isField && shortcuts.t) { //date=Today
			if (item.applyShortCut && item.applyShortCut(shortcuts, event)) {
				return true;
			}
		}
		if (shortcuts.f7 || shortcuts["?"] || shortcuts.s || shortcuts["."]) {
			syra_site.searcher && syra_site.searcher.gotoField(item.page);
			return true;
		}
		//if (event.shiftKey && shortcuts.right) {
		if (shortcuts.shift && shortcuts.right) {
			return _walkActionPanel(item, shortcuts, event);
		}
		if (shortcuts.left || shortcuts.right) {
			_walkPageTabs(item, shortcuts, event);
			return true;
		}
		if (shortcuts.tab || shortcuts.i || shortcuts.d) {
			if (item.page.externalAdapter && item.page.externalAdapter.applyShortCut) {
				if (item.page.externalAdapter.applyShortCut(item, shortcuts)) {
					return true;
				};
			}
		}
		if (menuStack.length > 0) {
			_clearMenuStack(item);
		}

	} else {
		if ((shortcuts.up || shortcuts.down || shortcuts.left || shortcuts.right) && (menuStack.length > 0)) {
			return _walkActionPanel(item, shortcuts, event);
		}
		if (shortcuts.enter) {
			if (isPage && item._lastItem) {
				_clearMenuStack(item);
				syra_menus.click.menuItem(item._lastItem, event);
				event.returnValue = false;
				event.preventDefault();
				event.stopPropagation();
				return true;
			}
		}
		if (shortcuts.enter || shortcuts.space) {
			if (isPage && item.isMessageBox) {
				item.onMenuClick({
					$bind: item.focusBtnId
				});
				event.returnValue = false;
				event.preventDefault();
				event.stopPropagation();
				return true;
			}
		}
		if (item.isField && item.page.isTranslationPage) {
			item.page.applyShortCut(shortcuts, item);
		}
		if (shortcuts.tab) {
			if (isPage && item.isMessageBox) {
				if (item.$itemButtons.length > 1) {
					var ii, len;
					for (ii = 0, len = item.$itemButtons.length; ii < len; ii++) {
						if (item.$itemButtons[ii].$bind == item.focusBtnId) {
							break;
						}
					}
					if (ii < len) {
						ii = event.shiftKey ? (ii == 0 ? len - 1 : ii - 1) : (ii + 1 == len ? 0 : ii + 1);
						syra_site.dom.toggleClass(item.getMenuItem(item.focusBtnId).domItem, "s-msgbox-button-default", false);
						item.focusBtnId = item.$itemButtons[ii].$bind;
						syra_site.dom.toggleClass(item.getMenuItem(item.focusBtnId).domItem, "s-msgbox-button-default", true);
						setTimeout(function() {
							item.getMenuItem(item.focusBtnId).domItem.focus();
						}, 5);
					}
				}
				event.returnValue = false;
				event.preventDefault();
				event.stopPropagation();
				return true;
			}
			if (item.isField && item.$item.$contentEditable) {
				syra_workingCopy.onInputChange(item, event);
				return true;
			}
		}
		if (item.page && item.page.externalAdapter && item.page.externalAdapter.applyShortCut) {
			if (item.page.externalAdapter.applyShortCut(item, shortcuts, event)) {
				return true;
			}
		}
		if (item.isField || item.isArrayField || item.isMenuGroupRoot) {
			if (item.isMenuGroupRoot && syra_menus.group.applyShortCut(item, shortcuts, event)) {
				return true;
			}
			if (item.applyShortCut) {
				if (item.applyShortCut(shortcuts, event))
					return true;
			}
			if (shortcuts.enter) {
				if (item.$item.$contentEditable) {
					syra_workingCopy.onInputChange(item, event);
					event.preventDefault();
					event.stopPropagation();
					return true;
				}
				if (item.$item.$bind === "$search" && item.articleParent.isArrayField) {
					var list = item.articleParent;
					if (list.isArrayField && list.searcher && list.searchNext) {
						list.searcher.onClick(list.searchNext.key, event);
					}
					return true;
				}

				// is ie10 the event on change is not sended on enter key when value change.
				var identity = syra_site.getBrowserIdentity();
				if (identity.isMSIE && item.input && identity.info.indexOf("10.0") != -1) {
					syra_workingCopy.onInputChange(item, event);
				}
				if (!item.page.isFusionPage && item.$item.$isFilterMode) {
					syra_workingCopy.onInputChange(item);
					return true;
				}
			}

		}
		if (item.isRecordArticle && !item.isFilterArticle) {
			if (shortcuts.insert) {
				return syra_menus.click.create(item);
			}
			if (shortcuts["delete"] && (shortcuts.shift || item.page.isFusionPage)) {
				//trigger delete action
				return syra_menus.click.deleteItem(item);
			}
		}

		if (shortcuts.space) {
			if (item.page.isNavigationPage && item.isMenuItem) {
				var record = item.articleParent;
				if (record && record.list && record.list.builder) {
					record.list.builder.clickBookmark(record);
				}
				event.preventDefault();
				event.stopPropagation();
				return true;
			}
		}
		if (isPage && item.shortcutSpyerList) {
			return item.shortcutSpyerList.applyShortCut(shortcuts, event);
		}
	}
};
