// syracuse-tablet/mockup/js/modals.js
"use strict";
(function($) {
	if (!$.smobile) $.smobile = {};
	var exports = $.smobile.modals = {};
	var _log = $.smobile.log;
	var helpers = $.smobile.helpers;

	exports.error = function(title, e, cb) {
		if (typeof e === "object" && e.fileName && e.stack) {
			e = helpers.toDiagnose({
				where: title,
				exception: e
			});
		} else if (typeof e === "string") {
			e = helpers.toDiagnose({
				where: title,
				message: e
			});
		} else {
			e = helpers.toDiagnose(e);
		}
		var body = JSON.stringify(e, null, 2);
		_log && _log("exports.displayError", body);
		exports.modal("error", {
			close: "Close",
			title: title || "Error",
			body: body
		}, cb);
	};
	exports.confirm = function(action, cb) {
		exports.modal("confirm", {
			title: "Confirm",
			action: action,
			yes: "yes",
			no: "no"
		}, function(action) {
			if (cb) cb("yes" === action);
		});
	};
	exports.modal = function(id, context, cb) {
		id = "modals_" + id;
		$.when($.stemplates.execSync(id, context)).always(function(html) {
			if (html == null) {
				alert("Modal not found - id=" + id + "\n" + JSON.stringify(context, null, 2));
				if (cb) cb("close");
			} else {
				var h = $(html).modal();
				var done = false;
				// Use <button data-dismiss"modal" for close action
				// Use <button data-action"myaction" for action other than close 
				h.on("hidden.bs.modal", function() {
					// Called when dialog is close - data-dismiss = "modal" or  h.modal('hide')
					h.unbind();
					h.remove();
					// Call cb only if no click action
					if (!done && cb) cb("close");
				});
				// For dialogs with option
				h.click("[data-action]", function(evt) {
					var t = $(evt.target);
					var act = t.attr("data-action");
					if (act) {
						done = true;
						evt.preventDefault();
						evt.stopPropagation();
						// Close modal
						h.modal('hide');
						if (cb) cb(act, t.attr("data-params"));
					}
				});
			}
		});
	};

	exports.init = function() {
		$.stemplates.preload(["modals/error", "modals/confirm"]);
	};

})(jQuery);
