// syracuse-ui-json/public/js/jsonTree.js
"use strict";

['/syracuse-ui-json/public/css/json-tree.css'].forEach(function(href) {
	$("<link/>", {
		rel: "stylesheet",
		type: "text/css",
		href: href,
	}).appendTo("head");
});

function fill(parent, val, visible) {
	if (!val || typeof val !== 'object') {
		$('<span>').text('' + val).appendTo(parent);
	} else if (Array.isArray(val)) {
		var $span = $('<span>').text(val.length ? '[...] ' + val.length + ' elements' : '[]').addClass('wf-json-length').appendTo(parent).click(function() {
			$ol.toggle();
		});
		var $ol = $('<ol>').addClass('wf-json-ol').appendTo(parent).toggle(visible);
		val.forEach(function(v) {
			var $li = $('<li>').appendTo($ol);
			fill($li, v, true);
		});
	} else {
		var keys = Object.keys(val);
		var $span = $('<span>').text(keys.length ? '{...}' : '{}').addClass('wf-json-length').appendTo(parent).click(function() {
			$ul.toggle();
		});
		var $ul = $('<ul>').addClass('wf-json-ul').appendTo(parent).toggle(visible);
		keys.forEach(function(k) {
			var $li = $('<li>').appendTo($ul);
			var $key = $('<div>').text(k).appendTo($li).addClass('wf-json-key');
			var $val = fill($li, val[k], false);
		});
	}
}

exports.create = function(div) {
	var value;
	return {
		setValue: function(val) {
			value = val;
			$(div).empty();
			fill($(div), val, true);
		},
		getValue: function() {
			return value;
		}
	};
};
