// syracuse-ui/lib/bookmarks/bookmarksProvider.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function BookmarksProvider(){
}

exports.BookmarksProvider = helpers.defineClass(BookmarksProvider, null, {




    _navigateRep: function(){
    
    },
    notifyDataChange: function(item, data){
        if (item) {
            switch (item.arrayLevel) {
                case "array":
                    this.items = helpers.object.clone(data, true);
                    //if(this.bookmarksPage) this.bookmarksPage.applyChange({"items":helpers.object.clone(data, true)});
                    this.updateSiteFavorites();
                    break;
                case "cell":
                    this.items[item.articleParent.$serverIndex][item.$item.$variantItemKey].title = data.title;
                    // if(this.bookmarksPage) this.bookmarksPage.applyChange({"items":helpers.object.clone(this.items, true)})
                    this.updateSiteFavorites();
                    break;
            }
        }
    },
    
    
    dispose: function(){
    
        this.items = this.map = this.clientMap = null;
    },
    initialize: function(){
    
    },
   
});
