// syracuse-ui/lib/site/spy/spyer.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

var _window, _spyedPage, _$prototype, _$dataset, _$receive, _$notify, _spyPage;

exports.getSpyedPage = function(){
    return _spyedPage;
};

exports.spyDataset = function($dataset, $receive){
    _$dataset = $dataset;
    _$receive = $receive;
    if (_spyPage) {
        _spyPage.spyDataset(_$dataset, _$receive);
    }
};
exports.setSpyedPage = function(page){
    (_spyedPage = page).isPageSpyed = true;
    _$prototype = page.$prototype
    if (_spyPage) {
        _spyPage.spyPrototype(_$prototype);
    }
    exports.spyDataset(page.dataset);
};

exports.onSpyPageLoaded = function(spyPage){
    _spyPage = spyPage;
    _spyPage.localize = document.site.localize;
    _spyPage.spyPrototype(_$prototype);
    _spyPage.spyDataset(_$dataset, _$receive);
    _spyPage.spyNotify(_$notify);
};

exports.spyNotify = function($notify){
    _$notify = $notify;
    if (_spyPage) {
        _spyPage.spyNotify(_$notify);
    }
};
exports.open = function(){
    _window = window.open("?representation=s-uitest-spy.$test", "_blank");
};

exports.dispose = function(){
    _window = _spyedPage = _$dataset = _$receive = _$notify = _spyPage = null;
};
