// syracuse-ui/lib/site/spy/spyGateway.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function SpyGateway() {}

exports.SpyGateway = helpers.defineClass(SpyGateway, null, {
	open: function() {
		this.spyWindow = window.open("?representation=s-uitest-spy.$test", "_blank");
	},
	setSpyedPage: function(page) {
		(this.spyedPage = page).isPageSpyed = true;
		this.spyPrototype(page.$prototype);
		this.spyDataset(page.dataset);
	},
	getSpyedPage: function() {
		return this.spyedPage;
	},
	spyPrototype: function($prototype) {
		this.$prototype = $prototype;
		if (this.spyPage) {
			this.spyPage.spyPrototype(this.$prototype);
		}
	},
	spyDataset: function($dataset, $receive) {
		this.$dataset = $dataset;
		this.$receive = $receive;
		if (this.spyPage) {
			this.spyPage.spyDataset(this.$dataset, this.$receive);
		}
	},
	spyNotify: function($notify) {
		this.$notify = $notify;
		if (this.spyPage) {
			this.spyPage.spyNotify(this.$notify);
		}
	},
	onSpyPageLoaded: function(spyPage) {
		this.spyPage = spyPage;
		this.spyPage.localize = document.site.localize;
		this.spyPage.spyPrototype(this.$prototype);
		this.spyPage.spyDataset(this.$dataset, this.$receive);
		this.spyPage.spyNotify(this.$notify);
	},
	dispose: function() {
		delete this.spyedPage;
		delete this.spyWindow;
		delete this.spyPage;
	}
});
