// syracuse-ui/lib/site/spy/spyPage.js
"use strict";
var helpers = require("syracuse-core/lib/helpers");
var DesktopPage = require("syracuse-ui/lib/page/desktopPage").DesktopPage;

function SpyPage() {}

exports.SpyPage = helpers.defineClass(SpyPage, DesktopPage, {
	loadBox: function($fieldsPrototype, initData) {
		var self = this;
		self.$prototype = {
			$title: "Survey Page",
			$properties: {
				$fieldPrototype: {
					"$title": "Prototype",
					"$type": "application/x-string",
					$rows: 50
				},
				$fieldNotify: {
					"$type": "application/x-string",
					$rows: 20
				},
				$fieldDataset: {
					"$title": "Dataset",
					"$type": "application/x-string",
					$rows: 50
				},
				$fieldReceive: {
					"$type": "application/x-string",
					$rows: 20
				}
			},
			$actions: {
				$notifyData: {
					$title: "notify"
				},
				$receiveData: {
					$title: "receive"
				}
			}
		};
		self.$item = {
			$layout: {
				$items: [{
					$layoutType: "row",
					$items: [{
						$bind: "$fieldPrototype",
						$isEditMode: true,
						$isAutoSizeDisabled: true,
						$skin: "s-json-field"
					}, {
						$bind: "$fieldDataset",
						$isEditMode: true,
						$isAutoSizeDisabled: true,
						$skin: "s-json-field"
					}, {
						$layoutType: "stack",
						$items: [{
							$category: "section",
							"$title": "notify",
							$layout: {
								$items: [{
									$skin: "s-mn-link",
									$category: "link",
									$bind: "$notifyData"
								}, {
									$bind: "$fieldNotify",
									isTitleHidden: true,
									$isEditMode: true,
									$isAutoSizeDisabled: true,
									$skin: "s-json-field"
								}]
							}

						}, {
							$category: "section",
							"$title": "receive",
							$layout: {
								$items: [{
									$skin: "s-mn-link",
									$category: "link",
									$bind: "$receiveData"
								}, {
									$bind: "$fieldReceive",
									isTitleHidden: true,
									$isEditMode: true,
									$isAutoSizeDisabled: true,
									$skin: "s-json-field"
								}]
							}

						}]

					}]
				}]
			}
		};
		DesktopPage.prototype.loadBox.call(self, initData);
		if (window.opener && window.opener.syra_site) {
			window.opener.syra_site.spyer.onSpyPageLoaded(this);
		}
	},
	notifyActionChange: function(menuItem, target, value) {
		if (value.$notifyData || value.$receiveData) {
			try {
				if (value.$receiveData) {
					window.opener.syra_site.spyer.getSpyedPage().applyChange(JSON.parse(this.boundFields.$fieldReceive[0].getDataValue()));
				}
				if (value.$notifyData) {
					window.opener.syra_site.spyer.getSpyedPage().notifyChangeToServer(JSON.parse(this.boundFields.$fieldNotify[0].getDataValue()));
				}
			} catch (error) {
				syra_site.showDiagnoses({
					$diagnoses: [{
						$message: error.message
					}]
				});
			}
			return;
		}
		DesktopPage.prototype.notifyActionChange.call(this, menuItem, target, value);
	},
	spyNotify: function($notify) {
		this.boundFields.$fieldNotify[0].setDataValue($notify ? JSON.stringify($notify, null, 2) : "");

	},
	spyPrototype: function($prototype) {
		this.boundFields.$fieldPrototype[0].setDataValue(JSON.stringify($prototype, null, 2));
		this.spyDataset(null);
		this.spyNotify(null);
	},
	spyDataset: function($dataset, $receive) {
		this.boundFields.$fieldDataset[0].setDataValue($dataset ? JSON.stringify($dataset, null, 2) : "");
		this.boundFields.$fieldReceive[0].setDataValue($receive ? JSON.stringify($receive, null, 2) : "");
	}
});
