// syracuse-soap/lib/server/genericOperations.js
"use strict";

exports.get = function(entity){
	
	
	var title = entity.charAt(0).toUpperCase() + entity.slice(1);
	var ns = entity + ":";
	var operations = {};

	
	// CREATE OPERATION
	var $create = {
		$title: "create"+title,
		$request: {
			$nokey: true,
			$type: "application/x-reference",
			$: {}
		},
		$response: {
			$nokey: true,
			$type: "application/x-reference",
			$: {
				$key: {
					$type: "application/x-string",
					$isMandatory: true
				}
			}
		}
	}
	$create.$request.$[entity] = {
		$type: ns + entity + "--type",
		$isMandatory: true
	}
	
	// READ OPERATION
	var $read = {
		$title: "read"+title,
		$request: {
			$nokey: true,
			$type: "application/x-reference",
			$: {
				$key: {
					$isMandatory: true,
					$type: "xs:string"
				},
				$binary: {
					$type: "xs:boolean",
					$default: "false"
				}
			}
		},
		$response: {
			$nokey: true,
			$type: "application/x-reference",
			$: {}
		}
	}
	$read.$response.$[entity] = {
		$type: ns + entity + "--type"
	}

	// UPDATE OPERATION
	var $update = {
		$title: "update"+title,
		$request: {
			$nokey: true,
			$type: "application/x-reference",
			$: {}
		},
		$response: {
			$nokey: true,
			$type: "application/x-reference",
			$: {
				$key: {
					$type: "application/x-string",
					$isMandatory: true
				}
			}
		}
	}
	$update.$request.$[entity] = {
		$type: ns + entity + "--type",
		$isMandatory: true
	}

	
	// DELETE OPERATION
	var $delete = {
		$title: "delete"+title,
		$request: {
			$nokey: true,
			$type: "application/x-reference",
			$: {
				$key: {
					$isMandatory: true,
					$type: "xs:string"
				}
			}
		},
		$response: {
			$nokey: true,
			$type: "application/x-reference",
			$: {
				deleted: {
					$type: "xs:boolean"
				}
			}
		}
	}

	
	// LIST OPERATION
	var $list = {
		$title: "list"+title,
		$request: {
			$nokey: true,
			$type: "application/x-reference",
			$: {
				$binary: {
					$type: "xs:boolean",
					$default: "false"
				},
				$count: {
					$type: "xs:integer",
					$default: "20"
				},
				$sdataWhere:{
					$type: "xs:string"
				},
				$orderBy:{
					$type: "tns:orderBy--list"
				},
				$pageInfo: {
					$type: "xs:string"
				}				
			}
		},
		$response: {
			$nokey: true,
			$type: "application/x-reference",
			$: {
				$itemsPerPage: {
					$type: "xs:integer"
				},
				$pageInfo: {
					$type: "xs:string"
				},
				$first: {
					$type: "xs:string"
				},
				$previous: {
					$type: "xs:string"
				},
				$next: {
					$type: "xs:string"
				},
				$last: {
					$type: "xs:string"
				},
				$totalResults: {
					$type: "xs:integer"
				}				
			}
		},
		$orderBy: {
			$nokey: true,
			$type: "application/x-reference",
			$: {
				$binding: {
					$type: "xs:string"
				},
				$descending: {
					$type: "xs:boolean",
					$default: "false"
				}
			}
		}
	}


	
	$list.$response.$[entity + "List"] = {
		$type: ns + entity + "--list"
	}
	
	
	
	return {
		$create: $create,
		$read: $read,
		$update: $update,
		$delete: $delete,
		$list: $list
	}

	
}
