// syracuse-ui/lib/diagnoses/diagnoseMessagePopup.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function DiagnoseMessagePopup() {}

exports.DiagnoseMessagePopup = helpers.defineClass(DiagnoseMessagePopup, null, {
	loadBox: function(elt, severity) {
		this.domItem = document.createElement("div");
		this.domItem.className = "s-diag-msg-popup";

		this.header = document.createElement("div");
		this.header.className = "s-diag-msg-popup-header";
		this.domItem.appendChild(this.header);

		this.body = document.createElement("div");
		this.body.className = "s-diag-msg-popup-body-" + severity;

		this._message = document.createElement("div");
		this._message.className = "s-diag-msg-popup-message";
		var html = document.createElement("div");
		html.innerHTML = elt.innerHTML;
		this._message.innerHTML = ($(html).find("span").contents().unwrap().end().end())[0].innerHTML;

		// appending
		this.body.appendChild(this._message);
		this.domItem.appendChild(this.body);
	},
	dispose: function() {
		this.domItem = this.header = this.body = this._message = null;
	}
});
