// syracuse-ui/lib/diagnoses/diagnosesPanel.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var DiagnoseMessagePopup = require('syracuse-ui/lib/diagnoses/diagnoseMessagePopup').DiagnoseMessagePopup;
var DiagnoseItem = require('syracuse-ui/lib/diagnoses/diagnoseItem').DiagnoseItem;
var _field = require('syracuse-ui/lib/diagnoses/diagnoseField');

function _addIconText(slot, text, severity) {
	slot.appendChild(syra_menus.addFontIconText(text, "s-diag-field s-diag-" + severity, severity == "info" ? severity : "diagnose"));
}

function DiagnosesPanel() {}

helpers.defineClass(DiagnosesPanel, null, {
	drawLinks: function(slot, message, page) {
		message.$diagnosePanelId = this.id;
		message.$diagnosesSlot = slot;
		syra_menus.setMenus(page || this.page, message, slot);
		delete message.$diagnosePanelId;
		delete message.$diagnosesSlot;
	},
	hasErrors: function(page) {
		for (var ii = this.fieldStack.length - 1; ii >= 0; ii--) {
			if (this.fieldStack[ii].isError()) {
				return true;
			}
		}
		return false;
	},
	onBoxItemInOut: function(box, onEnter) {
		if (box.severitiesCount && box.diagnoseBtn) {
			if (onEnter) {
				if (!box.diagnosePopupSlot) {
					box.diagnosePopupSlot = document.createElement("div");
					box.diagnosePopupSlot.className = "s-diag-field-slot-popup";
					var severities = ["fatal", "error", "warning", "info", "success"];
					for (var ii = 0, jj = severities.length; ii < jj; ii++) {
						var severity = severities[ii];
						if (box.severitiesCount[severity]) {
							_addIconText(box.diagnosePopupSlot, box.severitiesCount[severity] + " " + syra_local["box_" + severity], severity);
						}
					}
					box.diagnosePopupDiag = syra_site.dialogManager.openPopup(box.page, {
						content: {},
						slot: box.diagnosePopupSlot,
						position: {
							my: "right top",
							at: "right bottom",
							of: $(box.diagnoseBtn.link)
						},
						onClose: function() {
							syra_site.dom.removeChild(box.diagnosePopupSlot);
							box.diagnosePopupSlot = box.diagnosePopupDiag = null;
						}
					});
				}
			} else {
				box.diagnosePopupDiag && box.diagnosePopupDiag.close();
			}
		}
	},
	load: function(page) {
		this.page = page;
		this._diagnoseIndex = 0;
		this.id = page.id + "diagnosepanel";
		syra_store.add(this);
		this.domItem = document.createElement("div");
		this.domItem.className = "s-diagnose";
		this.page.diagnoseSlot.className = "s-diagnose-page-slot";
		this.domItem.syraItem = this.id;
		this.body = document.createElement("div");
		this.body.className = "s-diagnose-msg-list";
		this.domItem.appendChild(this.body);

		this.isCollapsed = true;
		this.clean();
		this.page.diagnoseSlot.appendChild(this.domItem);
		this._closeBtn = syra_menus.button.add({
			parent: this,
			slot: this.page.diagnoseSlot,
			text: syra_local.diag_close,
			iconOnly: true,
			css: "s-diagnose-close",
			btnclick: function() {
				this.parent.close();
			}
		});
	},
	clean: function(closing) {
		if (this.domItem) {
			syra_menus.button.display(this._expandBtn, false);
			syra_menus.button.display(this._nextBtn, false);
			syra_menus.button.display(this._closeBtn, false);
			syra_site.dom.display(this.domItem, false);
			if (this._menus && !this._menus.disposed) {
				this.removeItem(this._menus, true, true);
			}
			delete this._menus;
			if (closing) {
				while (this.stack.length) {
					this.stack[0].clean();
				}
			} else {
				syra_site.dom.empty(this.body);
				this.map = {};
				this.fieldStack = [];
				this.stack = [];
				this._focused = null;
			}
		}
	},
	getFieldItems: function(field) {
		var items = [];
		for (var ii = 0, jj = this.fieldStack.length; ii < jj; ii++) {
			if (this.fieldStack[ii].field == field) {
				items.push(this.fieldStack[ii]);
			}
		}
		return items;
	},
	onFieldItemInOut: function(field, onEnter) {
		_field.onFieldItemInOut(this, field, onEnter);
	},
	onMessageClick: function(event, btn) {
		var self = this;
		var id = btn.getAttribute("data-s-diagmsg-popup");
		if (id) {
			if (self.diagPopups && self.diagPopups[id]) {
				self.diagPopups[id].close();
			}
		} else {
			var label = btn.children[0];
			if (btn.children[1] && btn.children[1].className == "s-diagnose-msg-hide") {
				var msgPopup = new DiagnoseMessagePopup();
				msgPopup.loadBox(label, btn.getAttribute("data-s-severity"));
				var diagPopup = syra_site.dialogManager.openPopup(self.page, {
					content: msgPopup,
					slot: msgPopup.domItem,
					position: {
						my: "left top",
						at: "left bottom",
						of: $(label)
					},
					onClose: function() {
						msgPopup.dispose();
						btn.removeAttribute("data-s-diagmsg-popup");
						self.diagPopups[this.id] = null;
					},
					$isAutoClose: false
				});
				diagPopup.addCloseButton(msgPopup.header);
				btn.setAttribute("data-s-diagmsg-popup", diagPopup.id);
				self.diagPopups = self.diagPopups || {};
				self.diagPopups[diagPopup.id] = diagPopup;
			}
		}
	},

	close: function() {
		this.clean(true);
		this.page.resizeArticle();
	},
	showDiagnoses: function(message, page, options) {
		if (!message || (!message.$bind && !message.field && !message.$diagnoses)) {
			this.clean();
		} else {
			if (message.field && message.field.$item.$isExpressionChild) {
				return;
			}
			if (message.$links || message.$actions) {
				if (!this._diagViewLinks) {
					this._diagViewLinks = document.createElement("div");
					this._diagViewLinks.className = "s-diagnose-links-main";
					this.domItem.appendChild(this._diagViewLinks);
				}
				this.drawLinks(this._diagViewLinks, message, page);
			}
			if (message.field && (message.field.diagsSlot || message.field.mnPickers.diagnose || (message.$diagnoses === null))) {
				_field.clean(this, message.field);
			}
			if (message.$diagnoses) {
				for (var ii = 0, jj = message.$diagnoses.length; ii < jj; ii++) {
					var $diagnose = message.$diagnoses[ii];
					$diagnose.$severity = $diagnose.severity || $diagnose.$severity || "error";
					$diagnose.$message = $diagnose.message || $diagnose.$message;
					if ($diagnose.$message) {
						$diagnose.$message = syra_site.dom.formatHTMLText($diagnose.$message, true);
					}
				}
			}
			if (message.field) {
				_field.show(this, message);
			} else {
				if (message.$diagnoses && message.$diagnoses.length > 0) {
					for (var ii = 0, jj = message.$diagnoses.length; ii < jj; ii++) {
						this.addDiagnoseItem(page, message.$diagnoses[ii]).addTopLabel();
					}
				}
			}
			if (!options || !options.noViewer) {
				this.autoHide = message.autoHide;
				this.refresh();
			}
		}
	},
	addDiagnoseItem: function(page, $diagnose, field) {
		var item = new DiagnoseItem();
		item.id = this.id + (++this._diagnoseIndex);
		item.panel = this;
		item.page = page;
		item.$diagnose = $diagnose;
		this.map[item.id] = item;
		if (field) {
			item.article = field.articleParent;
			item.field = field;
			this.fieldStack.push(item);
		} else {
			item.article = page;
			this.stack.push(item);
		}
		return item;
	},
	walkToField: function(btn, field) {
		_field.walkToField(this, btn, field);
	},
	refresh: function() {
		var self = this;
		if ((self.page == syra_site) || (self.page.isPageLoaded && !self.page._isDataChanging)) {
			var count = self.stack.length + self.fieldStack.length;
			var showNext = self.fieldStack.length > 1;
			if (count) {
				if (showNext && !self._nextBtn) {
					self._nextBtn = syra_menus.button.add({
						parent: self,
						slot: self.page.diagnoseSlot,
						text: syra_local.diag_next,
						title: syra_local.diag_next_tooltip,
						css: "s-diagnose-nav-next",
						btnclick: function() {
							this.parent.walkToField(this.link);
						}
					});
				}

				for (var ii = 0, jj = self.stack.length; ii < jj; ii++) {
					self.stack[ii].showMessage(!self.isCollapsed);
				}
				for (var ii = 0, jj = self.fieldStack.length; ii < jj; ii++) {
					self.fieldStack[ii].showMessage(!self.isCollapsed);
				}
				if (!self._expandBtn) {
					self._expandBtn = syra_menus.button.add({
						parent: self,
						slot: self.page.diagnoseSlot,
						text: syra_local.diag_expand,
						css: "s-diagnose-expander",
						isHidden: true,
						btnclick: function() {
							var panel = this.parent;
							panel.isCollapsed = !panel.isCollapsed;
							panel.refresh();
							panel.page.resizeArticle();
						}
					});
				}
				syra_menus.button.display(self._expandBtn, count > 1);
				if (self.isCollapsed) {
					var item = self._focused || self.map[self.body.lastChild.syraId];
					item && item.showMessage(true);
					self._expandBtn.link.textContent = (count - 1) + syra_local.diag_expand;
				}
				self._expandBtn.link.textContent = self.isCollapsed ? ((count - 1) + " " + syra_local.diag_expand) : syra_local.diag_collapse;


				var zIndex = syra_site.dom.getTopZIndex() - 1;
				self._nextBtn && syra_site.dom.setZIndex(self._nextBtn.link, false, zIndex);
				self._expandBtn && syra_site.dom.setZIndex(self._expandBtn.link, false, zIndex);
				self._closeBtn && syra_site.dom.setZIndex(self._closeBtn.link, false, zIndex);
				syra_site.dom.setZIndex(self.domItem, false, zIndex);
			} else {
				syra_menus.button.display(self._expandBtn, false);
			}
			syra_menus.button.display(self._nextBtn, showNext);
			syra_menus.button.display(self._closeBtn, count > 0);
			syra_site.dom.display(self.domItem, count > 0);
			if (self.autoHide) {
				setTimeout(function() {
					self.clean();
					self.page && self.page.resizeArticle();
				}, self.autoHide || 4000);
				delete self.autoHide;
			}
		}
	},
	dispose: function() {
		syra_store.remove(this);
		syra_site.disposeObject(this);
	}
});

exports.showDiagnoses = function(message, page, options) {
	if (!page.diagnosesPanel) {
		page.diagnosesPanel = new DiagnosesPanel();
		page.diagnosesPanel.load(page);
	}
	page.diagnosesPanel.showDiagnoses(message, page, options);
};
