// syracuse-ui/lib/diagnoses/diagnoseItem.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

var _errors = ["fatal", "error", "warning"];

function DiagnoseItem() {}

exports.DiagnoseItem = helpers.defineClass(DiagnoseItem, null, {
	openDetail: function() {
		this.isExpanded = !this.isExpanded;
		this.topLabel.className = this.isExpanded ? "s-diagnose-msg-label" : "s-diagnose-msg-label-hidden";
		this._trace && syra_site.dom.display(this._trace, this.isExpanded);
	},
	showMessage: function(show) {
		syra_site.dom.display(this.msgItem, show);
		if (show && !this.isDetailChecked) {
			this.isDetailChecked = true;
			if (this.$diagnose.$stackTrace || this.topLabel.scrollHeight > this.topLabel.clientHeight) {
				var btn = syra_menus.addTextButton(syra_local.diag_stacktrace, "s-diagnose-msg-more", "openDetail");
				btn.syraId = this.id;
				this.msgItem.appendChild(btn);
			}
		}
	},
	addTopLabel: function() {
		this.topLabel = document.createElement("label");
		this.topLabel.className = "s-diagnose-msg-label-hidden";
		this.topLabel.textContent = this.$diagnose.$message;

		var severity = this.$diagnose.$severity;
		var msg = this.msgItem = document.createElement("div");
		msg.syraId = this.id;
		msg.className = "s-diagnose-msg s-diagnose-" + severity;
		msg.setAttribute("data-s-severity", severity);
		msg.syraOnClick = "onMessageClick";
		msg.appendChild(syra_menus.addFontIconFlag("s-diagnose-msg-flag s-diag-" + severity, severity == "info" ? severity : "diagnose"));

		msg.appendChild(this.topLabel);
		this.msgItem = msg;

		if (this.$diagnose.$links || this.$diagnose.$actions) {
			var slot = document.createElement("div");
			slot.className = "s-diagnose-links " + severity;
			this.panel.drawLinks(this.msgItem.appendChild(slot), this.$diagnose, this.page);
		}
		this.$diagnose.$stackTrace = this.$diagnose.$stackTrace || this.$diagnose.stackTrace;
		if (this.$diagnose.$stackTrace) {
			this.$diagnose.$stackTrace = syra_site.dom.formatHTMLText(this.$diagnose.$stackTrace, true);
			this._trace = document.createElement("div");
			this._trace.className = "s-diagnose-stacktrace";
			this._trace.innerHTML = this.$diagnose.$stackTrace = syra_site.dom.formatHTMLText(this.$diagnose.$stackTrace, true);
			this._trace.style.display = "none";
			msg.appendChild(this._trace);
		}
		this.panel.body.appendChild(msg);
	},
	isError: function() {
		return _errors.indexOf(this.$diagnose.$severity) >= 0;
	},
	clean: function() {
		if (this.msgItem) {
			syra_site.dom.removeChild(this.msgItem);
			if (this == this.panel.focused) {
				this.panel.focused = null;
			}
		}
		var stack = this.field ? this.panel.fieldStack : this.panel.stack;
		var index = stack.indexOf(this);
		index >= 0 && stack.splice(index, 1);
		delete this.panel.map[this.id];
	},
	focus: function() {
		if (this.panel._focused) {
			syra_site.dom.removeChild(this.panel._focused.focuseFlag);
			delete this.panel._focused.focuseFlag;
		}
		this.panel._focused = this;
		if (!this.focuseFlag) {
			this.focuseFlag = syra_menus.addFontIconFlag("s-diagnose-msg-focus", "right_arrow");
			this.topLabel.parentNode.insertBefore(this.focuseFlag, this.topLabel);
		}
		this.msgItem && syra_page.scrollToItem(syra_site, this.msgItem, this.panel.domItem, true);
	}
});
