// syracuse-ui/lib/tests/pages/ui-sql-designer/designer.js
"use strict";
var _data = require("./data");
exports.scenarios = {
	load: function() {
		return [{
			$title: "Scenario 1",
			run: function() {
				this.page.refreshData(_data.test1);
				var field = this.page.boundFields["ui-sql-designer"];
				//if (field.$isEditMode) {

				// this.assert(fields && fields.length == 1 && fields[0] == syra_quality.domQuery.field.focus());
				//}
			}
		}];
	}
};


exports.load = function(page, finish) {
	var $props = page.$prototype.$properties;
	$props["ui-sql-designer"] = {
		$title: "ui-sql-designer",
		$type: "application/x-sql-designer"
	};

	page.refreshData = function(data) {
		delete(this.dataset.$properties = this.dataset.$properties || {})["ui-sql-designer"];
		delete this.dataset["ui-sql-designer"];
		this.applyChange({
			$properties: {
				"ui-sql-designer": {}
			},
			"ui-sql-designer": data
		});
	};

	syra_item.ensureMediaTypes(page.$prototype, function() {
		page.addItem(page.leftCol, {
			$isEditMode: true,
			$bind: "ui-sql-designer"
		});
		finish(page);
	});

};
