// syracuse-tablet/html/js/helpers/scroller/nativeScroller.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var Base = require('syracuse-tablet/html/js/helpers/scroller/scrollerBase').Klass;

var _consts = {
	dirToAxis: {
		v: "y",
		h: "x"
	},
	dirToOtherAxis: {
		v: "x",
		h: "y"
	}
};

var _Klass = utils.defineClass(function NativeScroller($$elmt, options) {
	Base.call(this, $$elmt, options);
}, Base, {
	_setCSS: function(dir) {
		dir = dir || this.dir || "v";
		var $$parent = this.$$elmt.parent();
		var css = {};
		css["overflow-" + _consts.dirToAxis[dir]] = "auto"; // scrolling axis
		css["overflow-" + _consts.dirToOtherAxis[dir]] = "hidden"; // "unused" or fixed axis
		$$parent.css(css);
		// remove default touch action on all elements
		// would be better if doing this by loading a css file with the appropriate * {} selector instead of setting it on each element
		$("*").css({
			"touch-action": "auto",
			"-ms-touch-action": "auto"
		});
	},
	init: function(viewRect, dir) {
		Base.prototype.init.call(this, viewRect, dir);
		this._setCSS(this.dir);
	}
});

exports.Klass = _Klass;
