// syracuse-tablet/html/js/helpers/scroller/scrollerHelpers.js
"use strict";

var uiRect = require('syracuse-tablet/html/js/ui/rect');


var _scrollerId = 1;
var _ctor = {
	desktop: require('syracuse-tablet/html/js/helpers/scroller/hammerScroller'),
	native: require('syracuse-tablet/html/js/helpers/scroller/nativeScroller')
};

var _createSimpleScroller = function($$elmt) {
	var $$scrolls = $$elmt.find(".s-m-scroll-element");
	var i;
	var scrollers = {};

	for (i = 0; i < $$scrolls.length; i++) {
		_scrollerId++;

		var $$se = $($$scrolls[i]);
		var $$wrapper = $$se.parent();
		var dir = $$wrapper.hasClass("s-m-scroll-wrapper-h") ? "h" : "v";

		$$wrapper.attr("data-scroller-id", _scrollerId);

		var parentScroller = null;
		var $$parentScroller = $$wrapper.parent().closest(".s-m-scroll-wrapper");
		if ($$parentScroller.length > 0) {
			parentScroller = scrollers[$$parentScroller.attr("data-scroller-id")];
		}

		var css = {};
		css["overflow-" + dir] = "auto";
		$$wrapper.css(css);

		var gestureMgr = _newScroller($$se, {
			direction: dir,
			valMax: 0,
			name: "_scroller_" + _scrollerId,
			isPageScroller: false
		}, parentScroller);
		scrollers[_scrollerId] = gestureMgr;

		var scrollRect = uiRect.elmtRect($$se, "outer");
		var wrapperRect = uiRect.elmtRect($$wrapper, "outer");
		var viewRect = wrapperRect.intersectRect(scrollRect);
		if (viewRect && !viewRect.contains(scrollRect)) {
			gestureMgr.init(viewRect);
		} else {
			gestureMgr.reset();
		}
	}

	return scrollers;
};

var _newScroller = function($$elmt, options, parentMgr) {
	var md = new window.MobileDetect(window.navigator.userAgent);
	var name;
	if (md.mobile()) {
		name = "native";
	} else {
		name = "desktop";
	}
	var Klass = _ctor[name];
	if (!Klass) throw new Error("Scroller not initialized");

	return new Klass($$elmt, options, parentMgr);
};
exports.newScroller = _newScroller;
exports.createSimpleScroller = _createSimpleScroller;
exports.init = function() {
	_ctor.desktop = require('syracuse-tablet/html/js/helpers/scroller/hammerScroller').Klass;
	_ctor.native = require('syracuse-tablet/html/js/helpers/scroller/nativeScroller').Klass;
};
