// syracuse-tablet/html/js/authoring/arrays/authArrayGeneral.js
"use strict";
var authPropsGeneral = require('syracuse-tablet/html/js/authoring/authoringPropertiesGeneral');
var authPropsChart = require('syracuse-tablet/html/js/authoring/authoringPropertiesChart');
var locale = require('syracuse-tablet/html/js/helpers/locale');

var _templates = {
	subtitle: '\
		<section class="s-m-prop-subtitle" {{#if id}}id="{{id}}"{{/if}} style="{{#if notHidden}}overflow:hidden;{{/if}}white-space:nowrap;text-overflow:ellipsis;{{#if margin}}margin-bottom:{{margin}}px;{{/if}}">\
			{{#if title}}<p>{{title}}</p>{{/if}}\
			<div></div>\
		</section>',
	subPropTitle: ' \
		<div class="panel panel-primary s-m-panel-auth-prop s-m-auth-prop-type-{{type}}"> \
			<div style="overflow:hidden;" class="panel-heading" role="tab" id="heading_{{id}}"> \
				<a style="width:100%;display:table;" data-toggle="collapse" data-parent="#{{#if parentId}}{{parentId}}{{/if}}{{#unless parentId}}accordion{{/unless}}" href="#collapse_{{id}}" {{#if expanded}} aria-expanded="true"{{/if}} aria-controls="collapse_{{id}}" {{#unless expanded}}class="collapsed"{{/unless}}>\
					<div style="display:table-row;white-space: nowrap;">\
						<h4 class="panel-title" style="display:table-cell;"> \
							{{label}}\
						</h4> \
						{{#if info}}<span style="padding-left: 10px;font-size:x-small;display:table-cell;text-align: right;"><i>{{info}}</i></span>{{/if}}\
					</div>\
				</a> \
			</div> \
			<div id="collapse_{{id}}" class="panel-collapse collapse {{#if expanded}}in{{/if}}" role="tabpanel" aria-labelledby="heading_{{id}}"> \
			</div> \
		</div>',
	statusLine: '\
		<span style="display:table;table-layout:fixed;width:100%">\
			<span style="display:table-row;">\
				<span style="display:table-cell;overflow:hidden;text-overflow:ellipsis;white-space: nowrap;">{{field}} {{operator}} {{value}}</span>\
				<span class="{{css}}" style="{{style}}display:table-cell;width:{{sizeColor}}px;height:{{sizeColor}}px;margin-right:5px;"></span>\
			</span>\
		</span>'
};

var _getHtml = function(name, ctx) {
	var key = name + "compiled";
	var tmpl = _templates[key];
	if (!tmpl) {
		tmpl = _templates[name];
		tmpl = _templates[key] = Handlebars.compile(tmpl);
	}
	return tmpl(ctx);
};

var _properties = {
	"arrayDisplay": function(controller, sel, data) {
		var isCube = data.control.prototype.data("$cube");
		this.name = "arrayDisplay";
		this.label = null;
		this.key = "$display";
		this["default"] = "table";
		this.values = [{
			value: "table"
		}];
		if (data.control.prototype.isSingleArray()) {
			this.values.push({
				value: "separator"
			});
		} else if (!isCube) {
			this.values.push({
				value: "card"
			});
		}
		if (isCube) {
			this.values.push({
				value: "chart"
			});
		}
		this.createMarkup = authPropsGeneral.createMarkupRadio;
		this.refreshMarkup = authPropsGeneral.refreshMarkupRadio;
	},
	// only card
	"arraySeparator": {
		"key": "$separator",
		"default": "blank",
		values: [{
			value: "blank"
		}, {
			value: "dash"
		}, {
			value: "colon"
		}, {
			value: "semicolon"
		}],
		createMarkup: authPropsGeneral.createMarkupRadio,
		refreshMarkup: authPropsGeneral.refreshMarkupRadio
	},
	// only query array
	"arrayProtoFilters": function(controller, sel, data) {
		this.name = "arrayProtoFilters";
		this.label = null;
		this.key = "$filters";
		// No filters displayed by default for vignette
		this["default"] = controller.isVignette ? "none" : "list";
		this.values = [{
			value: "list"
		}, {
			value: "tabs"
		}, {
			value: "none"
		}];
		this.createMarkup = authPropsGeneral.createMarkupRadio;
		this.refreshMarkup = authPropsGeneral.refreshMarkupRadio;
	},

	// only query array - link to details values are created dynamically
	"arrayQueryLink": function(controller, sel, data) {
		this.name = "arrayQueryLink";
		this.label = null;
		this.isHidden = true;
		this.values = [];
		this.key = "$detailsLink";
		this["default"] = "$details";
		this.createMarkup = authPropsGeneral.createMarkupRadio;
		this.refreshMarkup = authPropsGeneral.refreshMarkupRadio;

		this.initMarkup = function(controller, sel, data) {
			var links = data.control.prototype.getDataByPath("$item.$links");
			if (!links || links.length === 0) return;
			if (controller.isVignette) {
				// Add the capability to open query full page - Default option
				this["default"] = "$queryfullpage";
				this.values.push({
					value: "$queryfullpage",
					label: locale.text("auth.arrayQueryLink.queryfullpage")
				});
			}
			var self = this;
			// Add $links
			$.each(links, function(key, link) {
				var method = (link && link.$method) || "GET";
				if (link && method === "GET") {
					// For system level links (starting with $, we only allow the ones we know they are supported right now)
					if (key.indexOf("$") !== 0 || ["$edit", "$details"].indexOf(key) > -1) {
						self.values.push({
							value: key,
							label: link.$title ? controller.prototype.resolveExpression(link.$title) + " - " + key : key
						});
					}
				}
			});
			this.isHidden = (this.values.length === 1 && this.values[0].value === "$details");
		};
	},

	// Options of array
	"arrayOptions": function(controller, sel, data) {
		this.name = "arrayOptions";
		this.label = null;
		this.key = "$arrayOptions";
		if (!data.control.isArrayField()) {
			this.values = [{
				"key": "showPagination",
				"default": controller.isVignette ? false : true
			}, {
				"key": "showSortFilter",
				"default": controller.isVignette ? false : true
			}, {
				"key": "showSearch",
				"default": controller.isVignette ? false : true
			}];
		} else {
			this.values = [];
		}
		if (exports.checkDisplay(sel, data, "table") && !data.control.prototype.isSingleArray()) {
			this.values.push({
				"key": "showRowDetailLink",
				// No row detail link by default
				"default": false
			});
		}
		this.values.push({
			"key": "rowIndex",
			"default": exports.checkDisplay(sel, data, "card")
		});
		if (exports.checkDisplay(sel, data, "card")) {
			this.values.push({
				"key": "hideEmptyRows",
				"default": false
			});
		}
		this.createMarkup = authPropsGeneral.createMarkupCheckboxes;
		this.refreshMarkup = authPropsGeneral.refreshMarkupCheckboxes;
	}
};

exports.getHtml = _getHtml;

exports.checkDisplay = function(sel, data, display) {
	var value = data.articleSelection ? data.articleSelection["$display"] : null;
	return display === (value || "table");
};
exports.initModuleProperties = function(dest) {
	authPropsGeneral.initProperties(_properties, dest);
};

exports.getModuleProperties = function($$elmt, controller, props, sel, data) {
	if (controller.state.type !== "chartdetail") {
		if (data.control.$type == "tablet/x-array-chart") {
			props.push(authPropsChart.getChartDetailButton(controller, sel, data));
		}
		props.push(_properties.arrayDisplay);
	}
	if (exports.checkDisplay(sel, data, "separator")) {
		props.push(_properties.arraySeparator);
	} else {
		props.push(_properties.arrayOptions);
		if (!data.control.isArrayField() && controller.$filtersGet() && controller.$filtersGet().length > 1) {
			props.push(_properties.arrayProtoFilters);
		}
		if (controller.prototype.isQuery()) {
			props.push(_properties.arrayQueryLink);
		}
	}
};
