// syracuse-ui/lib/field/menus/menusController.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var ContextMenu = require('./contextMenu').ContextMenu;

function MenusController(){
}

exports.MenusController = helpers.defineClass(MenusController, null, {
    load: function(field){
        this.field = field;
    },
    dispose: function(){
        document.controller.disposeObject(this.contextMenu);
        delete this.contextMenu;
        delete this.field;
    },
	 appendPicker: function(picker){
        this._ensureContextMenu(picker);
        this.contextMenu.appendPicker(picker);
    },
    _ensureContextMenu: function(){
        if (!this.contextMenu) {
            this.contextMenu = new ContextMenu();
            this.contextMenu.create(this.field, {
                $isBindDisabled: true
            });
        }
    },
    ensurePickerVisible: function(show){
        if (this.contextMenu && this.contextMenu.menuPicker) {
            this.contextMenu.menuPicker.style.display = show ? "" : "none";
        }
    },
    setMenus: function(metaData){
        this._applyChange(metaData.$links);
        this._applyChange(metaData.$actions, true);
    },
    _addToContextMenu: function($menu, $isAction, $bind, article){
        $menu.$bind = ($menu.$sourceBind = $bind) + "-" + this.field.id;
        this._ensureContextMenu();
        var bound = article.menuItems[$menu.$bind];
        if (bound) {
            bound[0].setMenu($menu);
        }
        else {
            $menu.$isAction = $isAction;
            this.contextMenu._menusBox.addMenuItem($menu);
        }
    },
    _applyChange: function($menus, $isAction){
        if ($menus === null) {
            if (this.contextMenu) {
                this.contextMenu.remove();
                delete this.contextMenu;
            }
        }
        else {
            if ($menus) {
                var self = this;
                var article = self.field.getArticle();
                var menusKeys = Object.keys($menus);
                if (menusKeys.length == 0) {
                    self._ensureContextMenu();
                }
                else {
                    menusKeys.forEach(function($bind){
                        switch ($bind) {
                            case "$details":
                                if (self.field && self.field.$isDetailLinkIgnore) {
                                    self._addToContextMenu($menus[$bind], $isAction, $bind, article);
                                }
                                break;
                            case "$tunnel":
                            case "$suggest":
                            case "$lookup":
                            case "$select":
                                break;
                            case "$lazyload":
                                self._ensureContextMenu();
                                break;
                            default:
                                self._addToContextMenu($menus[$bind], $isAction, $bind, article);
                                break;
                        }
                    });
                }
            }
        }
    }
});
