// syracuse-ui/lib/page/navigation/navigationItem/navigationMenuBlock.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var ChildField = require('syracuse-ui/lib/field/child/childField').ChildField;
var NavigationMenuBlockRecord = require('./navigationMenuBlockRecord').NavigationMenuBlockRecord;

function NavigationMenuBlock() {}

exports.NavigationMenuBlock = helpers.defineClass(NavigationMenuBlock, ChildField, {
	loadBox: function() {
		this.$skin = this.$item.$skin = "s-nav-block";
		this.ChildRecordClass = NavigationMenuBlockRecord;
		ChildField.prototype.loadBox.call(this);
		this.articleParent.domItem.className += " s-nav-block-item";
		this.articleParent.reorderPicker.className += " s-nav-block-reorder";
		this.onReorderItem(this.articleParent);
		this.isMenuBlock = true;
		this.ensureChildBlockStyle();

	},
	ensureChildBlockStyle: function() {
		var isChild = false;
		var articleParent = this.articleParent;
		while (articleParent && !isChild && articleParent != this.page) {
			if (articleParent.isMenuBlock) {
				isChild = true;
				break;
			}
			articleParent = articleParent.articleParent;
		}
		document.site.toggleClass(this.articleParent.domItem, "s-nav-block-item-child", isChild);
	},
	onReorderItem: function(record) {
		this.ensureChildBlockStyle();
		//document.site.toggleClass(record.domItem, "s-nav-items-item-block", record.domItem.parentNode.childElementCount >1);        
	},
	onItemEnterLeave: function(event, record) {
		// ChildField.prototype.onItemEnterLeave.call(this, event);
	}
});
