// syracuse-ui/lib/page/navigation/navigationItem/navigationMenuBlock.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var ChildField = require('syracuse-ui/lib/field/child/childField').ChildField;
var NavigationMenuBlockRecord = require('./navigationMenuBlockRecord').NavigationMenuBlockRecord;

function NavigationMenuBlock() {}

exports.NavigationMenuBlock = helpers.defineClass(NavigationMenuBlock, ChildField, {
	dispose: function() {
		this.subModuleRecord = null;
		ChildField.prototype.dispose.call(this);
	},
	loadBox: function() {
		this.$skin = this.$item.$skin = this.page.isAdminMode ? "s-nav-admin-block" : "s-nav-block";
		this.ChildRecordClass = NavigationMenuBlockRecord;
		this.subModuleRecord = this.articleParent.list.subModuleRecord;
		ChildField.prototype.loadBox.call(this);
		this.articleParent.domItem.className += "  " + this.$skin + "-item";
		if (this.articleParent.reorderBtn) {
			this.articleParent.reorderBtn.className += " s-nav-block-reorder";
		}
		this.onReorderItem(this.articleParent);
		this.isMenuBlock = true;
	},
	onReorderItem: function(record) {},
	onItemInOut: function() {
		//important for compatibility
	}
});
