// tools/git/pre-commit-rules.js
"use strict";
var fs = require('fs');
//var beautify = require('js-beautify/beautify').js_beautify;
var beautifyCount = 0;
var errCount = 0;

function error(msg) {
	console.error("ERROR: " + msg);
	errCount++;
}

function info(msg) {
	console.log(msg);
}

function acceptFile(f) {
	var ext = f.substring(f.lastIndexOf('.'));
	if (ext !== '.js' && ext !== '._js') return false;
	if (f.indexOf('node_modules/tools/') >= 0) return false;
	if (f.indexOf('syracuse-') < 0 && f.indexOf('/node_modules/tools/') < 0) return false;
	var start = f.indexOf("node_modules/");
	if (f.indexOf('/deps/') > 0 || f.indexOf('/node_modules/', start + 1) > 0) return false;
	if (f.indexOf('/html/mobile/') > 0 && f.indexOf('/html/mobile/js/debug') < 0) return false;
	if (f.indexOf('syracuse-ui/themes/desktop/sage') >= 0) return false;
	return true;
}

exports.processFileSync = function(file) {
	if (!acceptFile(file)) return;
	if (!fs.existsSync(file)) return; // file deleted by commit
	var contents = fs.readFileSync(file, 'utf8');
	if (!/^\uFEFF?["']use strict["'];/.test(contents.substring(0, contents.indexOf('\n')))) error(file + ': "use strict"; missing at top of file (must be first line)');
	if (/^[^\/\"\']*debugger/m.test(contents)) error(file + ': contains active "debugger" statement');
	if (/^(<<<<<<<|=======|>>>>>>>)/m.test(contents)) error(file + ': unresolved merge conflict (search for <<<<, ==== , >>>>)');
	// FOR NOW RETURN HERE - WILL ENABLE BEAUTIFY LATER
	return;
	var beautified = beautify(contents, {
		indent_char: '\t',
		indent_size: 1,
	});
	if (beautified !== contents) {
		info(file + ": beautified");
		beautifyCount++;
		fs.writeFileSync(file, beautified, 'utf8');
	}
}

exports.finish = function() {
	if (beautifyCount > 0) console.error(beautifyCount + " files have been beautified. Please run commit again!");
	if (errCount > 0) console.error(errCount + " errors detected by pre-commit hook. Aborting!")
	process.exit(errCount > 0 || beautifyCount > 0);

}
