// syracuse-ui/lib/page/soap/soapPage.js
"use strict";
var helpers = require("syracuse-core/lib/helpers");
var DesktopPage = require("syracuse-ui/lib/page/desktopPage").DesktopPage;

function SoapPage() {}

helpers.defineClass(SoapPage, DesktopPage, {
	loadBox: function(initData) {
		this.$responseCount = 0;
		this.$item.$layout.$layoutType = "tabs";
		this.$item.$layout.$items[0].$title = "Request";
		DesktopPage.prototype.loadBox.call(this, initData);
	},
	onMenuClick: function(menuItem) {
		if (menuItem.$bind == "$invoke") {
			var self = this;
			if (syra_form.validate(self)) {
				var invokeMenu = self.$prototype.$links.$invoke;
				syra_url.formatMenuUrl(self, menuItem, null, function($url) {
					menuItem.$url = $url;
					syra_ajax.send({
						page: self,
						url: $url,
						accept: menuItem.$type,
						contentType: menuItem.$contentType,
						data: self.sendBag,
						method: menuItem.$method,
						noDisplayErr: menuItem.noDisplayErr || false,
						success: function(data, response) {
							var newTab = self.layoutContent.createChildItem({
								$title: "Response " + (++self.$responseCount),
								$category: "section",
								$isCloseable: true,
								$isEmptyVisible: true,
								$layout: {
									$items: []
								}
							});
							var $itemPage = {
								layoutSlot: newTab.body,
								$category: "vignette",
								openerUrlSegments: syra_url.parse($url),
								$representation: response.data
							};
							$itemPage.openerUrlSegments.$facet = "$response";
							var initData = $itemPage.$representation.$prototype;
							var databody = initData.body;
							delete initData.body;
							syra_pageBuilder.load($itemPage, function(page) {
								page.masterPage = self;
								delete initData.$description;
								initData.body = databody;
								newTab.openTab();
								page.applyChange(initData);
							});
						}
					});
				});
			}
			return false; //not 
		}
		return true;
	}
});

exports.open = function(segments) {
	syra_ajax.get({
		url: segments.$url,
		success: function(data, response, requestUrl) {
			var $itemPage = {
				layoutSlot: document.createElement("div"),
				$isEditMode: true,
				$pageCategoryClass: SoapPage,
				$representation: data,
				openerUrlSegments: segments
			};
			$itemPage.openerUrlSegments.$facet = "$request";
			$itemPage.openerUrlSegments.fullUrl = document.location.href;
			syra_site.onMainPageChange($itemPage);
		}
	});
};
