// syracuse-ui/lib/page/soap/soapPage.js
"use strict";
var helpers = require('syracuse-core').helpers;
var DesktopPage = require("syracuse-ui/lib/page/desktopPage").DesktopPage;

function SoapPage() {}

helpers.defineClass(SoapPage, DesktopPage, {
	load: function(initData) {
		this.$responseCount = 0;
		this.$item.$layout.$layoutType = "tabs";
		this.$item.$layout.$items[0].$title = "Request";
		DesktopPage.prototype.load.call(this, initData);
	},
	onMenuClick: function(options) {
		if (options.menu.$bind == "$invoke") {
			var self = this;
			if (syra_form.validate(self)) {
				var invokeMenu = self.$prototype.$links.$invoke;
				syra_url.formatMenuUrl(self, options.menu, null, function($url) {
					options.menu.$url = $url;
					syra_ajax.send({
						page: self,
						url: $url,
						accept: options.menu.$type,
						contentType: options.menu.$contentType,
						data: self.sendBag,
						method: options.menu.$method,
						noDisplayErr: options.menu.noDisplayErr || false,
						success: function(data, response) {
							var newTab = self.layoutContent.createChildItem({
								$title: "Response " + (++self.$responseCount),
								$category: "section",
								$isCloseable: true,
								$isEmptyVisible: true,
								$layout: {
									$items: []
								}
							});
							var $itemPage = {
								layoutSlot: newTab.body,
								$category: "vignette",
								urlSeg: syra_url.parse($url),
								$representation: response.data
							};
							$itemPage.urlSeg.$facet = "$response";
							var initData = $itemPage.$representation.$prototype;
							var databody = initData.body;
							delete initData.body;
							syra_pageBuilder.load({
								$itemPage: $itemPage,
								success: function(page) {
									page.masterPage = self;
									delete initData.$description;
									initData.body = databody;
									newTab.openTab();
									page.applyChange(initData);
								}
							});
						}
					});
				});
			}
			return false; //not 
		}
		return true;
	}
});

exports.open = function(urlSeg) {
	syra_ajax.get({
		url: urlSeg.$url,
		success: function(data, response, requestUrl) {
			var $itemPage = {
				layoutSlot: document.createElement("div"),
				$isEditMode: true,
				$pageCategoryClass: SoapPage,
				$representation: data,
				urlSeg: urlSeg
			};
			urlSeg.$facet = "$request";
			urlSeg.fullUrl = document.location.href;
			syra_site.onMainPageChange($itemPage);
		}
	});
};
