// syracuse-ui/lib/mobile/article/articleMenusPage.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var localize = require('syracuse-ui/lib/mobile/localize');

function ArticleMenusPage(){
}

exports.ArticleMenusPage = helpers.defineClass(ArticleMenusPage, null, {
    _initialize: function(){
        var self = this;
        
        // page
        self.$$item = $("<div>").attr("data-role", "page").attr("data-theme", document.site.$item.$layoutOptions.dialogbox.theme.$page).appendTo(document.site.$$container);
        
        // header
        self.$$header = $("<div>").attr("data-role", "header").attr("data-theme", document.site.$item.$layoutOptions.dialogbox.theme.$header).appendTo(self.$$item);
        $("<h1/>").appendTo(self.$$header).text(localize.page.articleMenusPageTitle);
        
        // handling dialog close
        self.$$header.delegate("a", "click", function(){
            return self._close();
        });
        
        // core
        //self.$$core = $('<div/>').attr("data-role", "content").appendTo(self.$$menusPage);
        self.$$core = $('<div/>').appendTo(self.$$item);
        
        // boolean
        self.menusPageInitialized = true;
        
        // set transparent background
        self._setBackground();
    },
    show: function($item, boxParent){
        var self = this;
        
        // initialize if necessary
        if (!self.menusPageInitialized) {
            self._initialize();
        }
        
        // appending article menus
        self._appendMenus($item, boxParent);
        
        // show dialog menus page
        $.mobile.changePage(self.$$item, {
            transition: "none",
            role: "dialog",
            changeHash: false,
            reverse: false
        });
        
        // hack to avoid JQM close behaviour
        self.$$item.data("dialog").close = function(){
        }
    },
    _appendMenus: function($item, boxParent){
        var self = this;
        self.menus = document.itemFactory.load(self.$$core, $item, boxParent);
    },
    _close: function(){
        var self = this;
        setTimeout(function(){
            self._dispose();
            setTimeout(function(){
                $.mobile.changePage(document.site.mainPage.$$item, {
                    transition: "none",
                    changeHash: false
                });
            }, 200);
        }, 10);
        return false;
    },
    _setBackground: function(){
        //TO IMPROVE ? REMOVE ?
        // -- hack for transparent dialog background --
        /*
         $('div[data-role="dialog"]').live('pagebeforeshow', function(e, ui){
         ui.prevPage.addClass("ui-dialog-background ");
         });
         
         $('div[data-role="dialog"]').live('pagehide', function(e, ui){
         $(".ui-dialog-background ").removeClass("ui-dialog-background ");
         });
         */
        // ---------------------------------------------
    },
    _dispose: function(){
        var self = this;
        self.$$header.undelegate();
        document.controller.disposeObject(self.menus);
        delete self.menus;
        // SOMETHING TO DO TO REMOVE FROM THE DOM
        self.$$item.empty();
        self.menusPageInitialized = false;
    }
});
