// syracuse-ui/lib/mobile/article/articleMenusPage.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var localize = require('syracuse-ui/lib/mobile/localize');
var MobilePopup = require("syracuse-ui/lib/mobile/site/popup").MobilePopup;

function ArticleMenusPage(){
}

exports.ArticleMenusPage = helpers.defineClass(ArticleMenusPage, MobilePopup, {
    _initialize: function(){
        var self = this;
        
        // page
        self.$$item = $("<div>").attr("data-role", "page").attr("data-theme", document.site.$item.$layoutOptions.dialogbox.theme.$page).appendTo(document.site.$$container);
        
        // header
        self.$$header = $("<div>").attr("data-role", "header").attr("data-theme", document.site.$item.$layoutOptions.dialogbox.theme.$header).appendTo(self.$$item);
        $("<h1/>").appendTo(self.$$header).text(localize.page.articleMenusPageTitle);
        
        // handling dialog close
        self.$$header.delegate("a", "click", function(){
            self.close();
            return false;
        });
        
        // core
        //self.$$core = $('<div/>').attr("data-role", "content").appendTo(self.$$menusPage);
        self.$$core = $('<div/>').appendTo(self.$$item);
        
        // boolean
        self.menusPageInitialized = true;
    },
    show: function($item, boxParent){
        var self = this;
        
        // initialize if necessary
        if (!self.menusPageInitialized) {
            self._initialize();
        }
        
        // appending article menus
        self._appendMenus($item, boxParent);
        
        // show dialog menus page
        $.mobile.changePage(self.$$item, {
            transition: "none",
            role: "dialog",
            changeHash: false,
            reverse: false
        });
        
        // hack to avoid JQM close behaviour
        self.$$item.data("dialog").close = function(){
        }
        
        // register popup in registered popups stack
        MobilePopup.prototype.init.call(this, self);
    },
    _appendMenus: function($item, boxParent){
        var self = this;
        self.menus = document.itemFactory.load(self.$$core, $item, boxParent);
    },
    dispose: function(){
        this.$$header.undelegate();
        if (this._contextMenu) {
            document.controller.disposeObject(this._contextMenu);
            delete this._contextMenu;
        }
        //TODO REMOVE FROM THE DOM
        this.$$item.empty();
        this.menusPageInitialized = false;
    }
});
