// syracuse-ui/lib/mobile/article/box.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Box = require('syracuse-ui/lib/common/article/box').Box;

function MobileBox(){
}

exports.MobileBox = helpers.defineClass(MobileBox, Box, {
    loadBox: function(){
        this.$headerTag = "h3";
        Box.prototype.loadBox.call(this);
    },
    _ensureSkin: function(blockType){
        Box.prototype._ensureSkin.call(this, blockType);
        var $layoutOptions = this.getArticle().$layoutOptions || this.getPage().$layoutOptions;
        if ($layoutOptions) {
            this.$theme = $layoutOptions.theme["$" + blockType];
        }
    },
    openBox: function(open){
        if (!this.isjqmBoxInit) {
            if (!this.$theme) {
                debugger;
            }
            this.$theme = this.$theme ||
            {
                $header: "a",
                $content: "a"
            };
            /*if (this.$item.$title) {
             this.$item.$isBoxCollapsable = !this.$isTabLayout;
             }*/
            if (this.applyCollapsable(this.$item.$isBoxCollapsable)) {
                if (this.getPage().$isjqmPage) {
                    document.site.refreshJQMPlugin(this.$$item, "collapsible");
                }
            }
            this.isjqmBoxInit = true;
        }
        Box.prototype.openBox.call(this, open);
    },
    applyCollapsable: function($isBoxCollapsable){
        if ($isBoxCollapsable) {
            if (!this.$$collapsWrapper) {
                if (this.$collapseCss) {
                    this.$$item[0].className = this.$$item[0].className.replace(this.$collapseCss, "");
                }
                this.$$item.attr("data-role", "collapsible").attr("data-collapsed", this.$item.$opened === false);
                this.$$item.wrap(this.$$collapsWrapper = $("<div style='padding:5px 12px'/>"));
                //!!!important avoid ui-collapsible-heading maring-8px errror
                this.$$item.attr("data-theme", this.$theme.$header).attr("data-content-theme", this.$theme.$content);
            }
        }
        else {
            if (!this.$contentThemeCss && (this.$item && !this.$item.$isBodyTransparent)) {
                if (this.$$item[0]) 
                    this.$$item[0].className += (this.$contentThemeCss = " ui-body ui-body-" + this.$theme.$content);
            }
            
        }
        return $isBoxCollapsable;
    }
});
