// syracuse-ui/lib/mobile/article/dialogPage.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Page = require('syracuse-ui/lib/common/article/page').Page;
var localize = require('syracuse-ui/lib/mobile/localize');
var ArticleMenusPage = require('syracuse-ui/lib/mobile/article/articleMenusPage').ArticleMenusPage;

function DialogPage(){
}

exports.DialogPage = helpers.defineClass(DialogPage, Page, {
    loadBox: function(initData){
        var $representationUrl = this.$prototype ? this.$prototype.$representationUrl : null;
        Page.prototype.loadBox.call(this, initData);
    },
    applyActionLinkChange: function(resources, record){
        Page.prototype.applyActionLinkChange.call(this, resources, record);
        if (resources || record) {
            var self = this;
            self.menuBoxes.forEach(function(menus){
                menus.refreshListView();
            });
        }
    },
    _drawHeaderBar: function(){
        this._$$headerBar = $("<header/>");
        this._$$headerBar.attr("data-role", "header").attr("data-theme", this.$layoutOptions.theme.$headerBar);
        $("<h1/>").appendTo(this._$$headerBar);
        this._$$headerBar.appendTo(this.$$item);
		this._$$headerBar.delegate("a","click",function(){
			return this.onClose;
		});
    },
    drawBox: function(){
        this.$item.$title = this.$item.$title || "{$title}";
        this.$theme = this.$layoutOptions.theme;
        this.$skin = "s-mobile-page";
        if (this.$isMainPage !== false) {
            this.$$item = $("<div data-role='page'/>").attr("data-theme", this.$theme.$page).appendTo(this.$$container);
            this._drawHeaderBar();
            this.$$pageContent = $("<div/>").attr("data-role", "content").attr("data-theme", this.$theme.$content).appendTo(this.$$item);
        }
        else {
            this.$$item = this.$$container;
            this.$$pageContent = this.$$item;
        }
        if (!this.$item.$isTitleHidden) {
            this.appendHeader(this.$$pageContent);
            //            if (this.$prototype.$description) {
            //                this.renderDescription(this.$prototype.$description);
            //            }
        }
        this.$$body = $("<div/>").addClass(this.$skin + "-body").appendTo(this.$$pageContent);
        
        this._renderLayoutContent(this.$item);
    },
    
    appendHeader: function($$container, hideHeader){
        this.$$header = $("<header/>").addClass(this.$skin + "-head");
        this.$$title = $("<div/>").addClass(this.$skin + "-title").appendTo(this.$$header);
        if (this.$item.$isTitleHidden || hideHeader) {
            this.$$header[0].style.display = "none"; //used by porletfield
        }
        this.$$header.prependTo($$container || this.$$item);
        this.renderTitle(this.$item.$title || this.id);
        if (this.$item.$description) {
            this.renderDescription(this.$item.$description);
        }
    }
});
