// syracuse-ui/lib/mobile/article/page.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Page = require('syracuse-ui/lib/common/article/page').Page;
var localize = require('syracuse-ui/lib/mobile/localize');
var ArticleMenusPage = require('syracuse-ui/lib/mobile/article/articleMenusPage').ArticleMenusPage;

function MobilePage(){
}

exports.MobilePage = helpers.defineClass(MobilePage, Page, {
    loadBox: function(initData){
        var $representationUrl = this.$prototype ? this.$prototype.$representationUrl : null;
        if ($representationUrl && $representationUrl.indexOf("x3.erp.SUPERV.AUTILISM") >= 0) {
            var $list = this.$item.$layout.$items[0].$layout.$items[0];
            $list.$format = "cardList";
            $list.$cardsByRowCount = 3;
            $list.$cardview = {
                $layout: {
                    $layoutType: "side",
                    $items: [{
                        $category: "block",
                        $layoutType: "stack",
                        $items: [{
                            $bind: "PHOTO",
                            $isTitleHidden: true
                        }]
                    }, {
                        $layoutType: "stack",
                        $items: [{
                            $bind: "USR",
                            $isTitleHidden: true
                        }, {
                            $bind: "INTUSR",
                            $isTitleHidden: true
                        }]
                    }]
                }
            }
        }
        else {
            if ($representationUrl && $representationUrl.indexOf("x3.erp.SUPERV.COMPANYM") >= 0) {
                var $list = this.$item.$layout.$items[0].$layout.$items[0];
                $list.$format = "cardList";
                $list.$cardsByRowCount = 3;
                $list.$cardview = {
                    $layout: {
                        $layoutType: "side",
                        $items: [{
                            $category: "block",
                            $layoutType: "stack",
                            $items: [{
                                $bind: "LOGO",
                                $isTitleHidden: true
                            }]
                        }, {
                            $layoutType: "stack",
                            $items: [{
                                $bind: "CPY",
                                $isTitleHidden: true
                            }, {
                                $bind: "CPYSHO",
                                $isTitleHidden: true
                            }]
                        }]
                    }
                }
            }
        }
        switch (this.$pageCategory) {
            case "dashboard":
                this.$authoringSubType = "dashboard";
                this.$autoFetch = false;
                break;
            case "portlet":
                this.$isMainPage = false;
                if (this.$facet == "$demo") {
                    this.$autoFetch = false;
                }
                this.$item.$isTitleHidden = true;
                break;
        }
        Page.prototype.loadBox.call(this, initData);
    },
    applyActionLinkChange: function(resources, record){
        Page.prototype.applyActionLinkChange.call(this, resources, record);
        if (resources || record) {
            var self = this;
            self.menuBoxes.forEach(function(menus){
                menus.refreshListView();
            });
        }
    },
    _drawHeaderBar: function(){
        if (this.$pageCategory == "portlet") {
            return;
        }
        this._$$headerBar = $("<header/>");
        this._$$headerBar.attr("data-role", "header").attr("data-add-back-btn", false).attr("data-position", "fixed").attr("data-theme", this.$layoutOptions.theme.$headerBar);
        
        $("<h1/>").appendTo(this._$$headerBar);
        this._$$headerBar.appendTo(this.$$item);
        
        // Home button
        document.itemFactory.load(this._$$headerBar, {
            $bind: "$home",
            $category: "link",
            $theme: document.site.$item.$layoutOptions.site.theme.$home,
            $noText: true,
            $icon: {
                $name: "home"
            },
            $format: "button",
            $button: {
                $position: "right"
            }
        }, this);
        
        // back button only if not home page
        document.itemFactory.load(this._$$headerBar, {
            $bind: "$back",
            $category: "link",
            $theme: document.site.$item.$layoutOptions.site.theme.$home,
            $noText: true,
            $icon: {
                $name: "back"
            },
            $format: "button",
            $button: {
                $position: "left"
            }
        }, this).onMenuClick = function(){
            parent.history.back();
            return false;
        };
    },
    _drawFooterBar: function(){
        var self = this;
        
        // footer bar
        self._$$footerBar = $("<footer class='ui-bar'/>").attr("data-role", "footer").attr("data-position", "fixed").attr("data-theme", self.$layoutOptions.theme.$footerBar);
        
        self._$$footerCopyright = $("<div>").addClass("copyright").text(localize.footer.copyright).appendTo(self._$$footerBar);
        
        // appending footer bar
        self._$$footerBar.appendTo(self.$$item);
        
    },
    drawBox: function(){
        this.$item.$title = this.$item.$title || "{$title}";
        this.$theme = this.$layoutOptions.theme;
        this.$skin = "s-mobile-page";
        if (this.$isMainPage !== false) {
            this.$$item = $("<div data-role='page'/>").attr("data-theme", this.$layoutOptions.theme.$page).appendTo(this.$$container);
            this._drawHeaderBar();
            this.$$pageContent = $("<div/>").attr("data-role", "content").attr("data-theme", this.$layoutOptions.theme.$content).appendTo(this.$$item);
            this._drawFooterBar();
        }
        else {
            this.$$item = this.$$container;
            this.$$pageContent = this.$$item;
        }
        if (!this.$item.$isTitleHidden) {
            this.appendHeader(this.$$pageContent);
            //            if (this.$prototype.$description) {
            //                this.renderDescription(this.$prototype.$description);
            //            }
        }
        this.$$body = $("<div/>").addClass(this.$skin + "-body").appendTo(this.$$pageContent);
        
        this._renderLayoutContent(this.$item);
        this.appendArticleMenus();
        
        if (this.$dialogMode == "modal") {
            debugger;
        }
    },
    
    appendHeader: function($$container, hideHeader){
        this.$$header = $("<header/>").addClass(this.$skin + "-head");
        this.$$title = $("<div/>").addClass(this.$skin + "-title").appendTo(this.$$header);
        if (this.$item.$isTitleHidden || hideHeader) {
            this.$$header[0].style.display = "none"; //used by porletfield
        }
        this.$$header.prependTo($$container || this.$$item);
        this.renderTitle(this.$item.$title || this.id);
        if (this.$item.$description) {
            this.renderDescription(this.$item.$description);
        }
    },
    appendArticleMenus: function(){
        var self = this;
        
        if (self.$pageCategory == "dashboard") {
            return;
        }
        
        self.$facetMenus = self._getFacetMenus();
        if (self.$facetMenus) {
        
            // article actions menus button
            self._$$footerMenusBtn = $("<a>").attr("data-role", "button").attr("data-icon", document.site.$item.$icons.menusIcon).text(localize.footer.menus).appendTo(self._$$footerBar);
            
            
            // handling menus page opening
            self._$$footerMenusBtn.bind("click", function(){
                var menusPage = new ArticleMenusPage();
                menusPage.show({
                    $bind: "$linkBox",
                    $skin: self.$layoutOptions.menus,
                    $category: "links",
                    $format: "selectmenu",
                    $excludeBind: self.$facetMenus.$linkBox.$excludeBind
                }, self);
                /*
                 if (!self.menusPage){
                 self.menusPage = new ArticleMenusPage();
                 }
                 self.menusPage.show({
                 $bind: "$linkBox",
                 $skin: self.$layoutOptions.menus,
                 $category: "links",
                 $format: "selectmenu",
                 $excludeBind: self.$facetMenus.$linkBox.$excludeBind
                 }, self);
                 */
            });
            
            
            // loading footer main action button
            document.itemFactory.load(self._$$footerBar, {
                $category: "link",
                $format: "button",
                $icon: {
                    $name: document.site.$item.$icons.mainAction[self.$facetMenus.$mainAction]
                },
                $theme: document.site.$item.$layoutOptions.site.theme.$mainAction,
                $bind: self.$facetMenus.$mainAction
            }, self);
            
        }
    }
});
