// syracuse-collaboration/lib/entities/soapEntity.js
"use strict"

exports.entity = {
	$properties: {
		name: {
			$title: "Name",
			$isMandatory: true,
//			$propagate: function(_, instance, val) {
//
////				console.log("NAME="+val);
////				
////				console.log("App="+util.inspect(instance));
////				
//////				var inst = instance.entities(_);
//////				var array = inst.toArray(_);
//////				console.log("ARRAY="+util.inspect(array));
//////				for (var rel in array){
//////					var defRelName = array[rel].name(_);
//////					console.log("NAME="+defRelName);
//////				}
//			}
		},
		title: {
			$title: "Title",
			$isMandatory: true
		},
	},
	$relations: {
		representations: {
			$title: "Representations",
			$type: "soapRepresentations",
			$isChild: true,
			$select: {
				$title: "Representations",
				$type: "lookupRepresentation",
				$fieldMap: { name: "name", title: "title" },
				$parameters: "application={applicationName}&contract={contractName}&dataset={endpointName}&entity={name}"
			}
		},
	},
	$defaultOrder: [["name", true]]
}
