// syracuse-tablet/test/autoUI/lib/ut.js
"use strict";

var _webdriver = require('selenium-webdriver');
var _assert = require('selenium-webdriver/testing/assert');

exports.Driver = require("./driver");
exports.App = require("./app");
exports.Sel = require("./selectors");
exports.Dom = require("./dom");
exports.Config = require("../config").config;

exports.log = function(text) {
	if (exports.Config.log) {
		console.log(text);
	}
};

/**
 * options: {
 *    testPageUrl: sdata url of page to open after login
 *              E.g.: http://localhost:8124/mobile2/x3/erp/GX3APP/AQMDEVICE?representation=AQMCRUDM.$query 
 * }
 */
exports.defaultBeforeAndAfter = function(options) {
	before(function() {
		this.timeout(30000);
		return exports.login(options);
	});
	after(function() {
		if (exports.Config.dontQuit) {
			return;
		}
		var delay = exports.Config.quitDelay || 0;
		this.timeout(delay == null ? 2000 : (delay + 2000));
		return exports.Driver.quit(delay);
	});
};

/**
 * options: {
 *    testPageUrl: sdata url of page to open after login
 *              E.g.: http://localhost:8124/mobile2/x3/erp/GX3APP/AQMDEVICE?representation=AQMCRUDM.$query 
 * }
 */
exports.login = function(options) {
	return exports.App.loadApp(options)
		.then(function() {
			return exports.App.ensureLogin(options);
		});
};

exports.resolve = function() {
	var deferred = new _webdriver.promise.Deferred();
	deferred.fulfill.apply(deferred, arguments);
	return deferred.promise;
};

exports.Deferred = function() {
	return new _webdriver.promise.Deferred();
};

exports.sleep = function(time) {
	return exports.Driver.Browser().sleep(time);
};
exports.sleep$ = function(time) {
	return function() {
		return exports.Driver.Browser().sleep(time);
	};
};

exports.assert = function(subject) {
	return new _assert.Assertion(subject);
};
exports.assertEquals = function(a, b, text) {
	var asrt = new _assert.Assertion(a);
	return asrt.equals(b, text);
};

/** 
 * Assert that is always ok to signal this fact at the end of a process
 */
exports.assertOk = function(text) {
	var a = new _assert.Assertion(true);
	return a.equals(true, text);
};

/** 
 * Assert that is always failed to signal this fact at the end of a process
 */
exports.assertFail = function(text) {
	var a = new _assert.Assertion(false);
	a.equals(true, text);
};
//======================================================================================
//===06/12/2016============================= Ltr added Begin ===========================
//======================================================================================
/**
 * options: {
 *    testPageUrl: sdata url of page to open after login
 *              E.g.: testPageUrl: "{$hostPort}/mobile2/$client/$local/$dataset/mobileDashboard('GRAPHIC-SUBLEVEL')"
 *     page Wait:   data-s-ut-page-is-current="$graphicsDashboard.$mobileDashboard" 
 *     see the files:   authoringRequest.js or authoringStatistic.js
 *     testPageUrl is different the Page wait 
 * }
 */
exports.defaultBeforeAndAfterPageWait = function(options) {
	before(function() {
		this.timeout(30000);
		return exports.loginPageWait(options);
	});
	after(function() {
		if (exports.Config.dontQuit) {
			return;
		}
		var delay = exports.Config.quitDelay || 0;

		this.timeout(delay == null ? 2000 : (delay + 2000));
		return exports.Driver.quit(delay);
	});
};
/**
 * options: {
 *    testPageUrl: sdata url of page to open after login
 *             E.g.: testPageUrl: "{$hostPort}/mobile2/$client/$local/$dataset/mobileDashboard('GRAPHIC-SUBLEVEL')"
 *     page Wait:   data-s-ut-page-is-current="$graphicsDashboard.$mobileDashboard" 
 *     see the files:   authoringRequest.js or authoringStatistic.js
 *     testPageUrl is different the Page wait  
 * }
 */
exports.loginPageWait = function(options) {
	return exports.App.loadApp(options)
		.then(function() {
			return exports.App.ensureLoginPageWait(options);
		});
};
// == Login SageID screen =============================================
exports.loginSageId = function(options) {
	return exports.App.loadApp(options)
		.then(function() {
			return exports.App.ensureLoginSageId(options);
		});
};
//======================================================================================
//===06/12/2016============================= Ltr added End =============================
//======================================================================================
