// syracuse-ui/lib/common/article/layouts/authoring/dragDrop.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

var _localize = {
    emptyDropArea: "Drop area",
    emptyDropAreaOn: "You can drop it",
    moveItem: "move '{0}'",
    switchBlock: "swith '{0}' as block",
    switchSection: "swith '{0}' as section",
    moveBlock: "move '{0}' in new block",
    moveSection: "move '{0}' in new section",
    moveFields: "move fields of '{0}'",
    newSection: "New section",
    newBlock: "New block"
};

function DragDrop(){
}

exports.DragDrop = helpers.defineClass(DragDrop, null, {
    onEmptyLayoutArea: function(event){
        this.moveDragImage(event, null, true);
    },
    onEmptyEvent: function(event, layout){
        switch (event.type) {
            case "mouseup":
                var $action = this.getDropAction(event, layout.box.authoringNode);
                if ($action) {
                    if ($action.$appendNew) {
                        var newBox = this.sourceNode.authoringView.addNewBox({
                            parentBox: layout.box.authoringNode.item.boxParent,
                            $category: $action.$appendNew,
                            layout: layout
                        });
                        newBox.layoutContent.appendNewItem({
                            newItem: this.sourceNode.item
                        });
                    }
                    else {
                        if ($action.$switch) {
                            if ($action.$switch == "block" && this.sourceNode.getChildrenLevel() == "block") {
                                this.sourceNode.authoringView.moveContent({
                                    sourceItem: this.sourceNode.item,
                                    levelMoved: "block",
                                    layout: layout,
                                    doSelectItem: false
                                });
                                layout = null;
                                this.sourceNode.item.layoutParent.layoutAuthoring.deleteChilBox(this.sourceNode.item);
                            }
                            else {
                                this.sourceNode.item.switchCategory($action.$switch);
                            }
                        }
                        if (layout) {
                            layout.appendNewItem({
                                newItem: this.sourceNode.item
                            });
                        }
                    }
                    document.site.authoringView.isUpdated = true;
                }
                this.stop();
                delete document.site.DDAuthoring;
                break;
            case "mousemove":
                this.moveDragImage(event, layout.box.authoringNode);
                break;
            case "mouseleave":
                layout.empty.$$delete[0].style.visibility = "";
                layout.empty.$$drop.toggleClass("s-drop-over", false);
                layout.empty.$$dropText.text(_localize.emptyDropArea);
                break;
            case "mouseenter":
                if (this.getDropAction(event, layout.box.authoringNode)) {
                    layout.empty.$$delete[0].style.visibility = "hidden";
                    layout.empty.$$drop.toggleClass("s-drop-over", true);
                    layout.empty.$$dropText.text(_localize.emptyDropAreaOn);
                    this.moveDragCue(null); //hide
                }
                break;
        }
    },
    
    dropItem: function(){
        var targetItem = this.targetNode.item;
        if (this.$action.$appendNew) {
            var newBox = this.targetNode.authoringView.addNewBox({
                parentBox: targetItem.boxParent,
                $category: this.$action.$appendNew,
                targetItem: targetItem,
                action: this.$drag.$insert,
                layout: targetItem.layoutParent
            });
            newBox.layoutContent.appendNewItem({
                newItem: this.sourceNode.item
            });
        }
        else {
            if (this.$action.$moveContent) {
                this.sourceNode.authoringView.moveContent({
                    sourceItem: this.sourceNode.item,
                    targetItem: targetItem,
                    action: this.$drag.$insert
                });
            }
            else {
                if (this.$action.$switch) {
                    if (this.$action.$switch == "block" && this.sourceNode.getChildrenLevel() == "block") {
                        this.sourceNode.authoringView.moveContent({
                            sourceItem: this.sourceNode.item,
                            targetItem: targetItem,
                            action: this.$drag.$insert,
                            levelMoved: "block",
                            doSelectItem: false
                        });
						this.sourceNode.item.layoutParent.layoutAuthoring.deleteChilBox(this.sourceNode.item);
                        targetItem = null;
                    }
                    else {
                        this.sourceNode.item.switchCategory(this.$action.$switch);
                    }
                }
                if (targetItem) {
                    targetItem.layoutParent.appendNewItem({
                        newItem: this.sourceNode.item,
                        targetItem: targetItem,
                        action: this.$drag.$insert,
                        doSelectItem: true
                    });
                }
                
            }
            
        }
        document.site.authoringView.isUpdated = true;
    },
    start: function(sourceNode, $$boundary){
        var self = this;
        self.sourceNode = sourceNode;
        self.boundary = $$boundary.offset();
        self.boundary.right = self.boundary.left + $$boundary.width();
        self.boundary.bottom = self.boundary.top + $$boundary.height();
        $(document).bind("mouseup.syradragdrop", function(event){
            if (self.targetNode) {
                self.dropItem();
            }
            self.stop();
            delete document.site.DDAuthoring;
        }).bind("mousemove.syradragdrop", function(event){
            if (!event.target || event.target && event.target.className.indexOf("s-author-drop-cue") < 0) {
                self.moveDragImage(event, null, true);
            }
            event.preventDefault();
        });
    },
    _validateMoveContent: function(sourceItem, targetItem){
        var sourceBox = sourceItem;
        while (sourceBox && sourceBox != targetItem.boxParent) {
            sourceBox = sourceBox.boxParent;
        }
        return !sourceBox;
    },
    getDropAction: function(event, targetNode){
        var $action = null;
        if (this.sourceNode != targetNode) {
            var text;
            if (this.sourceNode.$authoringLevel != "field") {
                var boxParent = targetNode.item.boxParent;
                while (boxParent) {
                    if (this.sourceNode.item == boxParent) {
                        return null;
                    }
                    boxParent = boxParent.boxParent
                }
            }
            switch (this.sourceNode.$authoringLevel) {
                case "section":
                    switch (targetNode.$authoringLevel) {
                        case "article":
                            text = _localize.moveItem;
                            $action = {
                                $move: true
                            };
                            break;
                        case "section":
                            if (this._isEventOnDropEmpty(event)) {
                                text = _localize.switchBlock;
                                $action = {
                                    $switch: "block"
                                };
                            }
                            else {
                                text = _localize.moveItem;
                                $action = {
                                    $move: true
                                };
                            }
                            break;
                        case "block":
                            text = _localize.switchBlock;
                            $action = {
                                $switch: "block"
                            };
                            break;
                        case "field":
                            if (this._validateMoveContent(this.sourceNode.item, targetNode.item)) {
                                text = _localize.moveFields;
                                $action = {
                                    $moveContent: true
                                };
                            }
                            break;
                    }
                    break;
                case "block":
                    switch (targetNode.$authoringLevel) {
                        case "article":
                            if ((targetNode.getChildrenLevel() || "block") == "block") {
                                text = _localize.moveItem;
                                $action = {
                                    $move: true
                                };
                            }
                            else {
                                text = _localize.switchSection;
                                $action = {
                                    $switch: "section"
                                };
                            }
                            break;
                        case "section":
                            if (this._isEventOnDropEmpty(event)) {
                                text = _localize.moveItem;
                                $action = {
                                    $move: true
                                };
                            }
                            else {
                                text = _localize.switchSection;
                                $action = {
                                    $switch: "section"
                                };
                            }
                            break;
                        case "block":
                            text = _localize.moveItem;
                            $action = {
                                $move: true
                            };
                            break;
                        case "field":
                            if (this._validateMoveContent(this.sourceNode.item, targetNode.item)) {
                                text = _localize.moveFields;
                                $action = {
                                    $moveContent: true
                                };
                            }
                            break;
                    }
                    break
                case "field":
                    $action = {
                        $move: true
                    };
                    var $authoringLevel = targetNode.$authoringLevel;
                    if (this._isEventOnDropEmpty(event)) {
                        $authoringLevel = targetNode.getChildrenLevel() || "field";
                    }
                    if ($authoringLevel != "field") {
                        $action.$appendNew = $authoringLevel;
                        text = ($authoringLevel == "section") ? _localize.moveSection : _localize.moveBlock;
                    }
                    text = text || _localize.moveItem;
                    break;
            }
            if (text) {
                this._ensureDragImage();
                this._$$info.text(text.replace("{0}", this.sourceNode.item.getTitle()));
            }
        }
        return $action;
    },
    _calculateDropCuePosition: function(event, targetNode){
        var $layoutType = targetNode.item.layoutParent.$layout.$layoutType;
        var $$target = ($layoutType == "tabs") ? targetNode.item.$$header : targetNode.item.$$item;
        var offset = $$target.offset();
        this.$drag = {
            left: offset.left,
            top: offset.top,
            width: $$target.outerWidth(),
            height: $$target.outerHeight()
        };
        //calculateDropBoundary
        var xmargin = (this.$drag.width * 0.25);
        var ymargin = (this.$drag.height * 0.25);
        var $dropBoundary = {
            left: this.$drag.left + xmargin,
            right: this.$drag.left + this.$drag.width - xmargin,
            top: this.$drag.top + ymargin,
            bottom: this.$drag.top + this.$drag.height - ymargin
        };
        var $position = {
            top: this.$drag.top,
            left: this.$drag.left
        };
        var isAfter = false;
        if ($layoutType == "stack") {
            isAfter = event.pageY > $dropBoundary.bottom;
            $position.top = this.$drag.top + (isAfter ? this.$drag.height : 0);
            $position.height = "0.6em";
            $position.width = this.$drag.width + "px";
        }
        else {
            isAfter = event.pageX > $dropBoundary.right;
            $position.top -= 3;
            $position.left = this.$drag.left + (isAfter ? (this.$drag.width + 3) : (-3));
            $position.width = "0.6em";
            $position.height = this.$drag.height + 6 + "px";
        }
        this.$drag.$insert = isAfter ? "insertAfter" : "insertBefore";
        return $position;
    },
    moveDragCue: function($position){
        if (!this.$$dropCue) {
            this.$$dropCue = $("<div class='s-author-drop-cue'/>").appendTo(document.site.$$container)
        }
        this.$$dropCue[0].className = "s-author-drop-cue s-author-drop-cue-" + this.sourceNode.$authoringLevel;
        if ($position) {
            this.$$dropCue[0].style.display = "block";
            this.$$dropCue.css($position);
        }
        else {
            this.$$dropCue[0].style.display = "none";
        }
    },
    _ensureDragImage: function(){
        if (!this._$$info) {
            this._$$info = $("<div class='s-author-drag-image'/>").appendTo(document.site.$$container);
            this._$$info.text(this.sourceNode.item.getTitle());
        }
    },
    _isEventOnDropEmpty: function(event){
        return event.currentTarget.className.indexOf("s-author-empty") >= 0;
    },
    moveDragImage: function(event, targetNode, isOut){
        this.$action = null;
        if (targetNode) {
            this.$action = this.getDropAction(event, targetNode);
            if (this._isEventOnDropEmpty(event)) {
                targetNode = null
            }
        }
        var top = Math.max(event.pageY, this.boundary.top);
        var left = Math.max(event.pageX, this.boundary.left);
        top = Math.min(top, this.boundary.bottom);
        left = Math.min(left, this.boundary.right);
        
        var inBoundary = (event.pageX == left && event.pageY == top);
        this._ensureDragImage();
        if (!isOut && this.$action) {
            this._$$info.toggleClass("s-drag-ok", true);
            this.targetNode = targetNode;
        }
        else {
            if (isOut) {
                this._$$info.text(_localize.moveItem.replace("{0}", this.sourceNode.item.getTitle()));
            }
            this._$$info.toggleClass("s-drag-ok", false);
            this.targetNode = null;
        }
        this.moveDragCue(this.targetNode ? this._calculateDropCuePosition(event, this.targetNode) : null);
        
        this._$$info.show().css({
            top: top + 15 + "px",
            left: left + 15 + "px"
        });
    },
    stop: function(){
        document.controller.disposeObject(this);
    },
    dispose: function(){
        $(document).unbind('.syradragdrop');
        delete this.targetNode;
        if (this._$$info) {
            this._$$info.remove();
        }
        if (this.$$dropCue) {
            this.$$dropCue.remove();
        }
        delete this.sourceNode;
    }
});
