// syracuse-ui/lib/common/article/layouts/authoring/layoutSettings.js
exports.$columns = ["50,50", "25,75", "75,25", "38,62", "62,38", "25,25,50", "25,50,25", "50,25,25", "33,33-33", "25,25,25,25"];
exports.$composite = {
    "composite-3-1": {
        $title: "",
        $layoutType: "stack",
        $items: [{
            $layoutType: "columns",
            $layoutSubType: "50,50",
            $items: [{
                $layoutType: "stack",
                $items: []
            }, {
                $layoutType: "stack",
                $items: []
            }]
        }, {
            $layoutType: "tabs",
            $items: []
        }, {
            $layoutType: "columns",
            $layoutSubType: "50,50",
            $items: [{
                $layoutType: "stack",
                $items: []
            }, {
                $layoutType: "stack",
                $items: []
            }]
        }]
    },
    "composite-3-2": {
        $title: "",
        $layoutType: "stack",
        $items: [{
            $layoutType: "columns",
            $layoutSubType: "50,50",
            $items: [{
                $layoutType: "stack",
                $items: []
            }, {
                $layoutType: "stack",
                $items: []
            }]
        }, {
            $layoutType: "columns",
            $layoutSubType: "50,50",
            $items: [{
                $layoutType: "stack",
                $items: []
            }, {
                $layoutType: "stack",
                $items: []
            }]
        }, {
            $layoutType: "columns",
            $layoutSubType: "50,50",
            $items: [{
                $layoutType: "stack",
                $items: []
            }, {
                $layoutType: "stack",
                $items: []
            }]
        }]
    },
    "composite-4-1": {
        $layoutType: "stack",
        $items: [{
            $layoutType: "columns",
            $layoutSubType: "50,50",
            $items: [{
                $layoutType: "stack",
                $items: []
            }, {
                $layoutType: "stack",
                $items: []
            }]
        }, {
            $layoutType: "columns",
            $layoutSubType: "50,50",
            $items: [{
                $layoutType: "stack",
                $items: []
            }, {
                $layoutType: "stack",
                $items: []
            }]
        }, {
            $layoutType: "columns",
            $layoutSubType: "50,50",
            $items: [{
                $layoutType: "stack",
                $items: []
            }, {
                $layoutType: "stack",
                $items: []
            }]
        }, {
            $layoutType: "columns",
            $layoutSubType: "50,50",
            $items: [{
                $layoutType: "stack",
                $items: []
            }, {
                $layoutType: "stack",
                $items: []
            }]
        }]
    },
    "composite-1-3": {
        $layoutType: "columns",
        $layoutSubType: "50,50",
        $items: [{
            $layoutType: "stack",
            $items: []
        }, {
            $layoutType: "stack",
            $items: [{
                $layoutType: "columns",
                $layoutSubType: "50,50",
                $items: [{
                    $layoutType: "stack",
                    $items: []
                }, {
                    $layoutType: "stack",
                    $items: []
                }]
            }, {
                $layoutType: "columns",
                $layoutSubType: "50,50",
                $items: [{
                    $layoutType: "stack",
                    $items: []
                }, {
                    $layoutType: "stack",
                    $items: []
                }]
            }, {
                $layoutType: "columns",
                $layoutSubType: "50,50",
                $items: [{
                    $layoutType: "stack",
                    $items: []
                }, {
                    $layoutType: "stack",
                    $items: []
                }]
            }]
        }]
    }
};
