// syracuse-soap/lib/genericOperations.js
"use strict";

exports.get = function(entity){
	
	
	var title = entity.charAt(0).toUpperCase() + entity.slice(1);
	
	var operations = {};

	
	// CREATE OPERATION
	var $create = {
		$title: "create"+title,
		$request: {
			$nokey: true,
			$type: "application/x-reference",
			$: {}
		},
		$response: {
			$nokey: true,
			$type: "application/x-reference",
			$: {
				$key: {
					$type: "application/x-string",
					$isMandatory: true
				}
			}
		}
	}
	$create.$request.$[entity] = {
		$type: "tns:" + entity,
		$isMandatory: true
	}
	
	// READ OPERATION
	var $read = {
		$title: "read"+title,
		$request: {
			$nokey: true,
			$type: "application/x-reference",
			$: {
				$key: {
					$isMandatory: true,
					$type: "xs:string"
				},
				$binary: {
					$type: "xs:boolean",
					$default: "false"
				}
			}
		},
		$response: {
			$nokey: true,
			$type: "application/x-reference",
			$: {}
		}
	}
	$read.$response.$[entity] = {
		$type: "tns:" + entity
	}

	// UPDATE OPERATION
	var $update = {
		$title: "update"+title,
		$request: {
			$nokey: true,
			$type: "application/x-reference",
			$: {}
		},
		$response: {
			$nokey: true,
			$type: "application/x-reference",
			$: {
				$key: {
					$type: "application/x-string",
					$isMandatory: true
				}
			}
		}
	}
	$update.$request.$[entity] = {
		$type: "tns:" + entity,
		$isMandatory: true
	}

	
	// DELETE OPERATION
	var $delete = {
		$title: "delete"+title,
		$request: {
			$nokey: true,
			$type: "application/x-reference",
			$: {
				$key: {
					$isMandatory: true,
					$type: "xs:string"
				}
			}
		},
		$response: {
			$nokey: true,
			$type: "application/x-reference",
			$: {
				deleted: {
					$type: "xs:boolean"
				}
			}
		}
	}

	
	// LIST OPERATION
	var $list = {
		$title: "list"+title,
		$request: {
			$nokey: true,
			$type: "application/x-reference",
			$: {
				$binary: {
					$type: "xs:boolean",
					$default: "false"
				}
			}
		},
		$response: {
			$nokey: true,
			$type: "application/x-reference",
			$: {}
		}
	}
	$list.$response.$["ArrayOf" + title] = {
		$type: "application/x-collection",
		$item: {
			$type: "application/x-reference",
			$: {}
	    }
	}
	$list.$response.$["ArrayOf" + title].$item.$[entity] = {
		$type: "tns:" + entity
	}
	
	
	
	return {
		$create: $create,
		$read: $read,
		$update: $update,
		$delete: $delete,
		$list: $list
	}

	
}
