// syracuse-ui/lib/authoring/palette/toolsbar/arrayToolsBar.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var ToolsBar = require("syracuse-ui/lib/authoring/palette/toolsbar/toolsBar").ToolsBar;
var ColumnsToolbox = require("syracuse-ui/lib/authoring/palette/toolsbar/columnsToolbox").ColumnsToolbox;
function ArrayToolsBar(){
}

exports.ArrayToolsBar = helpers.defineClass(ArrayToolsBar, ToolsBar, {
    dispose: function(){
        if (this.columnsToolBox) {
            this.columnsToolBox.dispose();
        }
        this.columnsToolBox = this.graphToolBox = null;
        ToolsBar.prototype.dispose.call(this);
    },
    onEndChangeStep: function(){
        ToolsBar.prototype.onEndChangeStep.call(this);
        if (this.columnsToolBox) {
            this.columnsToolBox.toggle(true);
        }
    },
    _loadInsertLinks: function(){
        var $items = [];
        var binds = Object.keys(this.awContext.$prototype.$properties);
        if (binds.length > 0) {
            var $actions = {};
            for (var ii = 0, jj = binds.length; ii < jj; ii++) {
                var bind = binds[ii];
                if (bind.charAt(0) != "$") {
                    var $field = this.awContext.$prototype.$properties[bind];
                    if ($field && $field.$type) {
                        var actionId = "$insertAction" + bind;
                        var title = $field.$title ? this.awContext.targetPage.getLocalizeText($field.$title) : "";
                        $actions[actionId] = {
                            $title: (title == "") ? bind : title
                        };
                        $items.push({
                            $bind: actionId,
                            $fieldCategory: "field",
                            $fieldBind: bind,
                            $css: "s-aw-add-item",
                            $value: $field.$type.replace("application/x-", "").replace("/", "-")
                        });
                    }
                }
            }
            var $item = this.awContext.awArticle.$item;
            var $fields = this.awContext.authorPage.getItemSettings($item.$layout);
            for (var ii = 0, jj = $fields.length; ii < jj; ii++) {
                if ($fields[ii].$bind) {
                    var $action = $actions["$insertAction" + $fields[ii].$bind];
                    if ($action) {
                        $action.$isHidden = true;
                    }
                }
            }
            if ($item.$cards) {
                var $fields = this.awContext.authorPage.getItemSettings($item.$cards.$layout);
                for (var ii = 0, jj = $fields.length; ii < jj; ii++) {
                    if ($fields[ii].$bind) {
                        var $action = $actions["$insertAction" + $fields[ii].$bind];
                        if ($action) {
                            $action.$isHidden = true;
                        }
                    }
                }
            }
            this.applyChange({
                $actions: $actions
            });
        }
        return $items;
    },
    ensureArrayBoxVisibility: function(resize){
        var hideColumns = false, hideCards = false;
        if (this.awContext.isArrayField) {
            if (this.awContext.awArticle.$item.$format == "cards") {
                hideColumns = true;
            }
            else {
                if (this.awContext.awArticle.$item.$cards) {
                    switch (this.awContext.awArticle.$item.$cards.$position || "inline") {
                        case "popup":
                            hideColumns = true;
                            break;
                        default:
                            break;
                    }
                }
                else {
                    hideCards = true;
                }
            }
            hideCards = true;
            this._hideBox(this.columnsToolBox, hideColumns);
            this._hideBox(this.articleToolBox, hideCards);
            if (hideColumns && this.openedBox == this.columnsToolBox.toolbarBox) {
                this.collapseBox(this.articleToolBox.toolbarBox, true);
            }
            if (hideCards && this.openedBox == this.articleToolBox.toolbarBox) {
                this.collapseBox(this.columnsToolBox.toolbarBox, true);
            }
            if (resize !== false) {
                this.onWindowResize();
            }
        }
    },
    _createColumnsToolbox: function(){
        var box = this.createBox({
            $category: "section",
            $title: this.awContext.localize.aw_listColumns,
            $layout: {
                $items: []
            }
        });
        this.columnsToolBox = box.toolBox = new ColumnsToolbox();
        this.columnsToolBox.load(this.awContext, box);
    },
    _createGrapToolbox: function(){
        var chart = this.awContext.awArticle.chart;
        if (chart && chart.$authoringType == "cube$highCharts") {
            var $fieldCube = this.awContext.awArticle.$prototype.$cube;
            if ($fieldCube && $fieldCube.$measures) {
                var $items = [];
                var $itemCube = this.awContext.awArticle.$item.$cube || {};
                var $itemMeasures = $itemCube || {};
                var delta = {};
                var binds = Object.keys($fieldCube.$measures);
                for (var ii = 0, jj = binds.length; ii < jj; ii++) {
                    var bind = binds[ii];
                    var $measure = $fieldCube.$measures[bind];
                    var $field = chart.createChoices();
                    $field.$title = $measure.$title ? this.awContext.targetPage.getLocalizeText($measure.$title) : "";
                    var $toolBind = "$cube" + bind;
                    this.$prototype.$properties[$toolBind] = $field;
                    $items.push({
                        $category: "field",
                        $bind: $toolBind,
                        $measure: bind,
                        $icon: {
                            $mode: "icon",
                            $path: "graphics/s-graph-"
                        },
                        $choiceLayout: "3",
                        $isEditMode: true,
                        $skin: "s-aw-field"
                    });
                    delta[$toolBind] = ($itemMeasures[bind] ? $itemMeasures[bind].$style : null) || "area";
                }
                this.graphToolBox = this.createBox({
                    $category: "section",
                    $skin: "s-aw-list",
                    $title: this.awContext.localize.aw_graph,
                    $layout: {
                        $items: $items
                    }
                });
                this.applyChange(delta);
            }
        }
    },
    _createArticleToolbox: function(){
        ToolsBar.prototype._createArticleToolbox.call(this, this.awContext.localize.aw_listCard);
    },
    _createBoxes: function(){
        var $items = this._loadInsertLinks();
        this._createColumnsToolbox();
        this._createArticleToolbox();
        this._createGrapToolbox();
        this._createInsertToolbox($items);
    },
    _getDefaultBox: function(){
        return (!this.columnsToolBox.$isHidden) ? this.columnsToolBox : this.articleToolBox;
    },
    onNotifyDataChange: function(field, value){
        if (field.$item.$measure) {
            var $cube = this.awContext.awArticle.$item.$cube || {};
            $cube.$measures = $cube.$measures || {};
            ($cube.$measures[field.$item.$measure] = $cube.$measures[field.$item.$measure] || {}).$style = value;
            this.awContext.awArticle.awAddin.applyFieldDesignMetaDataChange({
                $cube: $cube
            });
        }
        ToolsBar.prototype.onNotifyDataChange.call(this, field, value);
    }
});
