// syracuse-ui/lib/utility/externalAdapter.js
"use strict";
exports.onBeforeUploadFiles = function(options) {
	options.doEvent();
};

exports.onGlobalClick = function(event) {
	return true;
};
exports.onFieldClickPicker = function(options) {
	options.doEvent();
};
exports.onGraphClickLink = function(options) {
	return true;
};
exports.getFieldEvalTitle = function(field) {
	return null;
};
exports.onSelectRecordEvent = function(options) {
	options.doEvent();
};
exports.onSortClickEvent = function(options) {
	options.doEvent();
};
exports.onSearchClickEvent = function(options) {
	options.doEvent();
};
exports.onBoxToggle = function(options) {
	options.doEvent();
};
exports.onBoxClick = function(options) {
	return true;
};
exports.onFieldEvent = function(options) {
	return options.doEvent();
};
exports.releaseMainPage = function(options) {
	options.doEvent();
};
exports.onMenuItemClick = function(options) {
	options.doEvent();
};
exports.setButtonValue = function(options) {
	options.doStatements(options);
};
exports.applyDesignStyle = function() {
	return true;
};
exports.sortArray = function() {
	return true;
};
exports.onFilterEvent = function(options) {
	options.doEvent();
};
exports.getPagerTotalRcdLabel = function(pagerList, nbRecords, defaultLabel) {
	return nbRecords + " " + (nbRecords > 1 ? defaultLabel[1] : defaultLabel[0]);
};
exports.onUploadSucceeded = function(options) {
	return true;
};
exports.onRemoveUploadedFile = function(options) {
	options.doEvent();
};
exports.onRemoveUploadedFileSucceeded = function(options) {
	return true;
};
