// syracuse-ui/lib/utility/log.js
"use strict";

exports.isDisabled = false;

exports.group = {
	start: function(key) {
		console.group(key);
	},
	end: function() {
		console.groupEnd();
	}
};

exports.object = function(key, object) {
	if (!exports.isDisabled) {
		try {
			console.groupCollapsed(key);
			console.log("%c" + JSON.stringify(object, null, 2), "color:blue;");
			console.groupEnd();
		} catch (error) {

		}
	}
};

exports.page = {
	data: {
		startChange: function() {
			console.groupCollapsed("page.startChange");
		},
		applyChange: function(newData) {
			newData &&
				exports.object("page.applyChange", newData);
		},
		endChange: function() {
			console.groupEnd();
		},
	},
	load: {
		start: function(page, $itemPage) {
			console.group("pageBuilder.load " + page.constructor.name);
			var keys = ["$authorUrl", "$facet", "$isEditMode", "$pageCategory", "$pageCollaborationUrl", "$authorUrl"];
			for (var ii = 0, jj = keys.length; ii < jj; ii++) {
				var key = keys[ii];
				if (page[key] !== undefined) {
					if (typeof page[key] === "object") {
						console.log("%c" + key + ": \t" + JSON.stringify(page[key], null, 2), "color:blue;");
					} else {
						console.log("%c" + key + ": \t" + page[key], "color:blue;");
					}
				}
			}
			exports.object("openerUrlSegments", page.openerUrlSegments);
			exports.object("$prototype", page.$prototype);
			exports.object("$preferences", page.$preferences);
		},
		end: function() {
			console.groupEnd();
		}
	},
	loadBox: {
		start: function(page) {
			console.groupCollapsed("page.loadBox" + page.constructor.name);
		},
		end: function() {
			console.groupEnd();
		}
	}
};

exports.widgets = {
	layout: {
		load: function(layout) {
			console.groupCollapsed("layout.loadBox $layoutType:" + layout.$item.$layoutType);
		},
		end: function() {
			console.groupEnd();
		}
	},
	addItem: function($item, item) {
		var key = "widgets.addItem";
		if ($item.$bind) {
			key += " $bind: " + $item.$bind;
		}
		if (item) {
			key += " " + item.constructor.name;
			if (item.constructor.name == "MenuBlockField") {
				return; // circular
			}

		}
		try {
			exports.object(key, $item);
		} catch (error) {

		}

	}
};
