// syracuse-ui/lib/utility/ui-util.js
"use strict";

exports.clone = function(obj) {
	return syra_site.clone(obj);
};

exports.registerFlashWidget = function(context) {
	context.field.layoutSlot.className += " s-field-flash-slot";
	var boxParent = (context.field.attachedField || context.field).boxParent;
	while (boxParent) {
		boxParent.hasFlashChildField = true;
		boxParent = boxParent.boxParent;
	}
};


exports.parseExpression = function(context, expression, dataset) {
	return syra_expression.parse(context.field.articleParent, expression, dataset);
};



exports.fetch = function(context, options, success, error) {
	options = options || {};
	if (options.$url) {
		options.$url = exports.parseExpression(context, options.$url);
	}
	context.field.page.fetch(options, success, error);
};

exports.menu = {
	fire: function(context, $link, dataset) {
		$link = syra_site.clone($link);
		$link.$url = exports.parseExpression(context, $link.$url, dataset);
		syra_router.executeMenu($link, context.field.articleParent);
	}
};

exports.preference = {
	profile: {
		get: function(key) {
			return syra_preference.profile.get(key);
		},
		save: function(key, value) {
			syra_preference.profile.save(key, value);
		}
	},
	page: {
		get: function(context) {
			return syra_preference.page.getField(context.field);
		},
		save: function(context, value) {
			syra_preference.page.saveField(context.field, value);
		}
	}
};
