// sky-automation/scripts/hibernate.js
#!/usr/bin/env node

'use strict';

var config = require('config');
var AWS = require('aws-sdk');

AWS.config.update(config.aws);
var ec2 = new AWS.EC2({apiVersion: '2014-06-15'});

exports.createInstanceTags = function(instanceId, instanceTags) {
  var params = {
    Resources: [
      instanceId
    ],
    Tags: [
      {
        Key: 'start',
        Value: instanceTags[0]
      },
      {
        Key: 'stop',
        Value: instanceTags[1]
      },
      {
        Key: 'daystolive',
        Value: instanceTags[2]
      },
    ],
  };

	ec2.createTags(params, function(err, data) {
	  if (err)
      console.log(err, err.stack);
	  /*else
      console.log("Tags created successfully");*/
	});
}

exports.getInstanceIdFromIP = function(instanceIP, callback) {
  var params = {
    Filters: [
      {
        Name: 'private-ip-address',
        Values: [
          instanceIP,
        ]
      },
    ],
  };
  ec2.describeInstances(params, function(err, data) {
    var id = "";
    if (err) 
      console.log(err, err.stack);
    else if (data && data.Reservations.length > 0) {
        var reservations = data.Reservations;
        var instances = reservations[0].Instances;
        var instance = instances[0];
        id = instance.InstanceId;
        callback(id)
    }
  });
}

exports.validateTags = function(instanceTags) {
  var valid = validateTimeFormat(instanceTags[0]);
  if (!valid)
    throw new Error("Invalid 24 hour format - Start Time");

  valid = validateTimeFormat(instanceTags[1]);
  if (!valid)
    throw new Error("Invalid 24 hour format - End Time");

  validateStartEnd(instanceTags[0], instanceTags[1]);

  if (typeof instanceTags[2] !== "number" && Number(instanceTags[2]) === "NaN")
    throw new Error("Invalid value, not a number");

  if (Number(instanceTags[2]) > 100)
    throw new Error("Invalid value, too large")
}

function validateTimeFormat(timeStr) {
  return (timeStr.search(/^\d{2}:\d{2}$/) != -1) &&
            (timeStr.substr(0,2) >= 0 && timeStr.substr(0,2) <= 24) &&
            (timeStr.substr(3,2) >= 0 && timeStr.substr(3,2) <= 59);
}

function validateStartEnd(startTime, endTime) {
  var startHour = startTime.substr(0,2);
  var startMinute = startTime.substr(3, 2);
  var endHour = endTime.substr(0,2);
  var endMinute = endTime.substr(3, 2);

  var startTimeObject = new Date();
  startTimeObject.setHours(startHour, startMinute, 0);

  var endTimeObject = new Date();
  endTimeObject.setHours(endHour, endMinute, 0);
  if (startTimeObject > endTimeObject)
    throw new Error("Invalid start and end times, start time after end time");
}
