// sky-automation/scripts/hibernate.js
#!/usr/bin/env node

'use strict';

var config = require('config');
var AWS = require('aws-sdk');

AWS.config.update(config.aws);
var ec2 = new AWS.EC2({apiVersion: '2014-06-15'});

exports.createInstanceTags = function(instanceId, instanceTags) {
  var params = {
    Resources: [
      instanceId
    ],
    Tags: [
      {
        Key: 'start',
        Value: instanceTags[0]
      },
      {
        Key: 'stop',
        Value: instanceTags[1]
      },
      {
        Key: 'daymonthstart',
        Value: instanceTags[2]
      },
      {
        Key: 'daymonthend',
        Value: instanceTags[3]
      },
      {
        Key: 'dayweek',
        Value: instanceTags[4]
      },
    ],
  };

	ec2.createTags(params, function(err, data) {
	  if (err)
      console.log(err, err.stack);
	  /*else
      console.log("Tags created successfully");*/
	});
}

exports.getInstanceIdFromIP = function(instanceIP, callback) {
  var params = {
    Filters: [
      {
        Name: 'private-ip-address',
        Values: [
          instanceIP,
        ]
      },
    ],
  };
  ec2.describeInstances(params, function(err, data) {
    var id = "";
    if (err) 
      console.log(err, err.stack);
    else if (data && data.Reservations.length > 0) {
        var reservations = data.Reservations;
        var instances = reservations[0].Instances;
        var instance = instances[0];
        id = instance.InstanceId;
        callback(id)
    }
  });
}
