// syracuse-ui/lib/page/navigation/module/moduleList.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var DesktopList = require("syracuse-ui/lib/field/array/list").DesktopList;
var ModuleBuilder = require("./moduleBuilder").ModuleBuilder;

function ModuleList() {

}

exports.ModuleList = helpers.defineClass(ModuleList, DesktopList, {
	loadBox: function(initData) {
		this.emptyMessage = this.localize.dskpNavigationPageNoModule;
		this.$confirmDelete = this.localize.dskpNavigationPageConfirmDeleteModule;
		this.$isEditMode = false;
		this.BuilderClass = ModuleBuilder;
		this.$isQuickDesignerEnabled = false;
		this.$item.$isPagerHidden = true;
		this.$item.$selectMode = "row";
		this.selectFirstRecordByDefault = true;
		DesktopList.prototype.loadBox.call(this, initData);
	},
	_createMenusBox: function() {
		if (!this.isMenuLoaded) {
			this.isMenuLoaded = true;
			this.page.loadNewItem(this.menusSlot, {
				$layoutType: "row",
				$autoSize: true,
				$items: [{
					$skin: "s-list-menus-link",
					$bind: "$create",
					$category: "link",
					$icon: {
						$mode: "icon",
						$path: "list/s-list-"
					}
				}, {
					$skin: "s-list-menus-link",
					$bind: "$select",
					$category: "link",
					$icon: {
						$mode: "icon",
						$path: "list/s-list-"
					}
				}]
			}, this);
		}
	},
	dispose: function() {
		document.site.emptyDom(this._core);
		DesktopList.prototype.dispose.call(this);
	}
});
