// syracuse-ui/lib/page/navigation/module/moduleRecord.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var RecordArticle = require("syracuse-ui/lib/field/array/recordArticle").RecordArticle;


function ModuleRecord() {}

exports.ModuleRecord = helpers.defineClass(ModuleRecord, RecordArticle, {
	loadBox: function() {
		this.isModuleRecord = true;
		this.$defaultSkinSection = this.$defaultSkinBlock = this.list.$item.$skinCard || "s-h3";
		RecordArticle.prototype.loadBox.call(this);
	},

	reorderItem: function(targetRecord, isAfter) {
		if (targetRecord) {
			targetRecord.domItem.parentNode.insertBefore(this.domItem, isAfter ? targetRecord.domItem.nextSibling : targetRecord.domItem);
		} else {
			if (isAfter) {
				this.builder.body.appendChild(this.domItem);
			} else {
				this.builder.body.insertBefore(this.domItem, this.builder.body.firstChild);
			}
		}
	},
	remove: function() {
		if (this.$$item) {
			this.$$item.remove();
		}
	},
	applyReorderCapability: function(reorder) {
		if (this.hasReorderCapability = reorder) {
			if (!this.reorderPicker) {
				this.reorderPicker = document.createElement("div");
				this.reorderPicker.className = this.$skin + "-card-reorder";
				this.reorderPicker.syraReorderRecordId = this.list.id;
				document.site.ddManager.setDragSpot(this.reorderPicker, true);
				this.domItem.insertBefore(this.reorderPicker, this.domItem.firstChild);
			}
		} else {
			if (this.reorderPicker) {
				document.site.removeDomChild(this.reorderPicker);
				delete this.reorderPicker;
			}
		}
	},
	applyChange: function(metaData) {
		RecordArticle.prototype.applyChange.call(this, metaData);
		if (this.dataset.description && this.dataset.description != this.dataset.title) {
			this.domItem.title = this.dataset.description;
		}
	},
	drawBox: function() {
		this.$isVerticalDirection = true;
		this.dataRow = this.domItem = document.createElement("div");
		this.$skin = "s-nav-module";
		this.domItem.className = this.$skin;
		this.domItem.setAttribute("data-s-record", this.$uuid);
		this.$$item = $(this.domItem);
		this.dataSlot = this.body = document.createElement("div");
		this.body.className = this.$skin + "-title";
		this.domItem.appendChild(this.body);

		this._menusSlot = document.createElement("div");
		this._menusSlot.className = this.$skin + "-links";
		this.domItem.appendChild(this._menusSlot);
		this.applyReorderCapability(true);

		if (this.list.selector.isRowMode) {
			this.body.setAttribute("data-s-picker", "list-selector");
			this.body.className += " s-list-selector-row";
			this.body.setAttribute("data-s-record", this.$uuid);
		}

		this.cardItem = this.page.loadNewItem(this.body, {
			$bind: "title",
			$isMenusHidden: true,
			$isDetailLinkDisabled: true,
			$category: "field",
			$isEditMode: this.$isEditMode,
			$inplace: true
		}, this);
		this.list._core.appendChild(this.domItem);
	},
	_createMenusBox: function() {
		this.page.loadNewItem(this._menusSlot, {
			$noText: true,
			$category: "link",
			$skin: this.$skin + "-link",
			$bind: "$edit",
			$icon: {
				$mode: "icon",
				$path: "list/s-list-"
			}
		}, this);
		this.page.loadNewItem(this._menusSlot, {
			$noText: true,
			$category: "link",
			$skin: this.$skin + "-link",
			$bind: "$delete",
			$icon: {
				$mode: "icon",
				$path: "list/s-list-"
			}
		}, this);
	},
	onSelectMouseEvent: function(event) {
		document.site.toggleClass(this.domItem, "s-record-over", event.type == "mouseenter");
	},
	onModuleSelected: function(module) {
		this.modules[module.$uuid];
		if (module.subModules) {
			for (var ii = 0, jj = module.subModules.length; ii < jj; ii++) {

			}
		}
	},
	highlightSelection: function(selected) {
		document.site.toggleClass(this.domItem, "s-list-record-selected", selected);
		if (selected) {
			if (!this.subModulesItem) {
				this.subModulesItemSlot = document.createElement("div");
				this.subModulesItemSlot.className = "s-nav-submodules-slot";
				this.page.layoutContent.domItem.appendChild(this.subModulesItemSlot);
				this.subModulesItem = this.page.loadNewItem(this.subModulesItemSlot, {
					$isTitleHidden: true,
					$bind: "submodules"
				}, this);
			}
			this.subModulesItemSlot.style.display = "";
		} else {
			if (this.subModulesItemSlot) {
				this.subModulesItemSlot.style.display = "none";
			}
		}
	},
	dispose: function() {
		this.subModulesItem = this.subModulesItemSlot = null;
		this.cardItem = this.domItem = this.dataSlot = this.body = this._menusSlot = this.dataRow = null;
		RecordArticle.prototype.dispose.call(this);
	}
});
