// syracuse-ui/lib/page/navigation/module/moduleBar.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var BarSplitter = require("syracuse-ui/lib/page/bar/barSplitter").BarSplitter;

function ModuleBar(){
}

exports.ModuleBar = helpers.defineClass(ModuleBar, BarSplitter, {
    load: function(page){
        this.options = {
            $isResizerDisabled: true,
            $skins: {
                bar: "s-nav-modulebar"
            }
        };
        this.barSlot = document.createElement("div");
        this.barSlot.className = "s-nav-modulebar-slot";
        this.barBody = document.createElement("div");
        this.barBody.className = "s-nav-modulebar-body";
        this.barBody.setAttribute("data-s-inout", page.id);
        this.isSlotVisible = true;
        page.domItem.insertBefore(this.barSlot, page.domItem.firstChild);
        BarSplitter.prototype.load.call(this, page);
    },
    onBarBodyAdded: function(){
        this.page.loadNewItem(this.barBody, {
            $isTitleHidden: true,
            $bind: "modules"
        });
    },
    getArticle: function(){
        return this.page.getArticle();
    }
});
