// syracuse-ui/lib/page/navigation/module/moduleBar.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var BarSplitter = require("syracuse-ui/lib/page/bar/barSplitter").BarSplitter;

function ModuleBar() {}

exports.ModuleBar = helpers.defineClass(ModuleBar, BarSplitter, {
	load: function(page) {
		if (document.site.isTableDevice()) {
			this.options = {
				resizeDirection: {
					left: true
				},
				$skins: {
					bar: "s-nav-modulebar-tablet",
					openerPicker: "s-nav-modulebar-tablet-opener",
					resizePicker: "s-nav-modulebar-tablet-resizer"
				}
			};
		} else {
			this.options = {
				$isResizerDisabled: true,
				$skins: {
					bar: "s-nav-modulebar"
				}
			};
		}
		this.barSlot = document.createElement("div");
		this.barSlot.className = this.options.$skins.bar + "-slot";
		this.barBody = document.createElement("div");
		this.barBody.className = this.options.$skins.bar + "-body";
		this.barBody.setAttribute("data-s-inout", page.id);
		this.isSlotVisible = true;
		page.domItem.insertBefore(this.barSlot, page.domItem.firstChild);
		BarSplitter.prototype.load.call(this, page);
	},
	onBarBodyAdded: function() {
		this.page.loadNewItem(this.barBody, {
			$isTitleHidden: true,
			$bind: "modules"
		});
	},
	getArticle: function() {
		return this.page.getArticle();
	}
});
