// syracuse-ui/lib/page/navigation/module/moduleList.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var ModuleRecord = require("./moduleRecord").ModuleRecord;
var MultiList = require("syracuse-ui/lib/field/array/multiList").MultiList;

function ModuleList() {

}

exports.ModuleList = helpers.defineClass(ModuleList, MultiList, {
	loadBox: function(initData) {
		this.emptyMessage = this.localize.dskpNavigationPageNoModule;
		this.$confirmDelete = this.localize.dskpNavigationPageConfirmDeleteModule;
		this.$isEditMode = false;
		this.$isQuickDesignerEnabled = false;
		this.$item.$isPagerHidden = true;
		this.$item.$selectMode = "row";
		MultiList.prototype.loadBox.call(this, initData);
	},
	drawBox: function() {
		MultiList.prototype.drawBox.call(this);
		this.body = this._core;
		this.fieldValue.style.overflow = "auto";
		this.$skin = this.$item.$skin || "s-cards";
		this.applyDesignMetaData(this.$item, false);
		this.loadSelector();
		this.applyCapabilities();
		document.site.ddManager.toggleColResizerObserver(this, true);
		this.ensureLayoutMode();
		this.layoutSlot.appendChild(this.domItem);
	},
	initializeList: function() {
		this.$item.$isTopLabelAlignment = true;
		this.RecordClass = ModuleRecord;
		if (!this.$item.$layout || !this.$item.$layout.$items) {
			this.$item.$layout = this.defineDefaultCard().$layout;
		}
		MultiList.prototype.initializeList.call(this);
	},
	setDataBind: function(dataRecordSet, parentDataRecord, metaData) {
		MultiList.prototype.setDataBind.call(this, dataRecordSet, parentDataRecord, metaData);
		if (this.records.length && !this.selector.records.length) {
			this.selector.select(undefined, true);
		}
	},
	_createMenusBox: function() {
		if (!this.isMenuLoaded) {
			this.isMenuLoaded = true;
			this.page.ensureDesignedItemVisibility(this);
			this.page.loadNewItem(this.menusSlot, {
				$layoutType: "row",
				$autoSize: true,
				$items: [{
					$skin: "s-list-menus-link",
					$bind: "$create",
					$category: "link",
					$icon: {
						$mode: "icon",
						$path: "list/s-list-"
					}
				}, {
					$skin: "s-list-menus-link",
					$bind: "$select",
					$category: "link",
					$icon: {
						$mode: "icon",
						$path: "list/s-list-"
					}
				}]
			}, this);
		}
	},
	onWindowResize: function() {
		if (!this.disposed && this.records) {
			for (var ii = 0, jj = this.records.length; ii < jj; ii++) {
				this.records[ii].resizeResizableChildren();
			}
		}
	},
	isDraggable: function(target, event) {
		return this.isReorderRecordDraggable(target, event);
	},
	dispose: function() {
		document.site.emptyDom(this._core);
		this.rows = null;
		MultiList.prototype.dispose.call(this);
	}
});
