// syracuse-ui/lib/jobs/setting/jobsViewerPrototype.js
"use strict";
exports.getPrototype = function() {
	var localize = syra_site.localize;
	return {
		$properties: {
			$jobType: {
				$type: "application/x-choice",
				$title: localize.jobs_list,
				$value: {
					$type: "application/x-integer",
					$enum: []
				},
				$capabilities: "sort,filter"
			},
			$jobDetails: {
				"$type": "application/x-array",
				$item: {
					$properties: {
						jobTitle: {
							$title: localize.jobs_job_title,
							$type: "application/x-string",
							$capabilities: "sort,filter,alphaTab"
						},
						phase: {
							$type: "application/x-string",
							$title: localize.jobs_job_phase,
							$capabilities: "sort,filter,alphaTab"
						},
						phaseDetail: {
							$type: "application/x-string",
							$title: localize.jobs_job_phase_detail,
							$capabilities: "sort,filter,alphaTab"
						},
						progress: {
							$title: localize.jobs_progress,
							$type: "application/x-quantity",
							$value: {
								$title: "{$value} {$unit}",
								$type: "application/x-decimal",
								$format: "0.00"
							},
							$unit: "percentage",
							$capabilities: "sort,filter,alphaTab"
						},
						percentage: {
							"$title": "Percentage",
							"$type": "application/x-choice",
							"$value": {
								"$type": "application/x-integer",
								"$enum": [{
									"$value": 1,
									"$title": localize.jobs_percent
								}]
							},
							$capabilities: "sort,filter,alphaTab"
						},
						elapsedSeconds: {
							$type: "application/x-integer",
							$title: localize.jobs_elapsed_seconds,
							$capabilities: "sort,filter,alphaTab"
						},
						remainingSeconds: {
							$type: "application/x-integer",
							$title: localize.jobs_remaining_seconds,
							$capabilities: "sort,filter,alphaTab"
						},
						pollingMillis: {
							$type: "application/x-integer",
							$title: localize.jobs_polling_millis,
							$capabilities: "sort,filter,alphaTab"
						},
						progressIcon: {
							$type: "application/x-string",
							$title: localize.jobs_progress_icon
						},
						jobDiagnoses: {
							$type: "application/x-array",
							$item: {
								$type: "application/json",
								$properties: {
									diagMsg: {
										$type: "application/x-string"
									},
									diagStack: {
										$type: "application/x-array",
										$title: localize.jobs_diagStackTrace,
										$item: {
											$properties: {
												stackTrace: {
													$type: "application/x-string"
												}
											},
											$isEditMode: false,
											$isTitleRowHidden: true,
											$isTitleHidden: true,
											$isPagerHidden: true,
											$isMenuRecordHidden: true
										}
									}
								}
							}
						},
						startDate: {
							$type: 'application/x-date',
							$title: localize.jobs_job_startdate
						},
						startTime: {
							$type: 'application/x-time',
							$title: localize.jobs_job_starttime
						}

					},
					$links: {
						$clearJob: {
							$title: localize.jobs_clear_job
						}
					}
				}
			}
		},
		$links: {
			$clearList: {
				$title: localize.jobs_clear_list
			}
		}
	};

};
