// syracuse-ui/lib/jobs/setting/jobsViewerArticle.js
"use strict";
exports.getItem = function() {
	var $skin = "s-jobs-viewer";
	var localize = document.site.localize;
	return {
		$category: "section",
		$skin: $skin + "-h1",
		$isTitleHidden: true,
		$layout: {
			$items: [{
				$layoutType: "row",
				$autoSize: true,
				$items: [{
					$css: $skin + "-types-list",
					$category: "field",
					$bind: "$jobType",
					$isEditMode: true,
					$format: "$combo",
					$isTopLabelAlignment: false
				}, {
					$category: "link",
					$bind: "$clearList",
					$css: $skin + "-clear-list"
				}]
			}, {
				$skin: $skin + "-head",
				$css: $skin + "-head",
				$format: "cards",
				$bind: "$jobDetails",
				$isTitleHidden: true,
				$isPagerHidden: true,
				$isQuickDesignerEnabled: false,
				$isMenuRecordHidden: true,
				$cards: {
					$layout: {
						$items: [{
							$layoutType: "row",
							"$isFixedWidth": true,
							"$widths": "*,*,*,12",
							$items: [{
								$layoutType: "stack",
								$items: [{
									$bind: "$jobPrintIcon",
									$isTitleHidden: true,
									$category: "link",
									$noText: true,
									$skin: "s-job-view-type-print"
								}, {
									$bind: "$jobOperationIcon",
									$isTitleHidden: true,
									$category: "link",
									$noText: true,
									$skin: "s-job-view-type-operation"
								}]
							}, {
								$layoutType: "stack",
								$css: "s-job-view-phase-info",
								$items: [{
									$bind: "$progressOn",
									$isTitleHidden: true,
									$category: "link",
									$noText: true,
									$skin: "s-job-view-progress-value"
								}, {
									$bind: "$progressOff",
									$isTitleHidden: true,
									$category: "link",
									$noText: true,
									$skin: "s-job-view-progress-value-dis"
								}, {
									$bind: "progress",
									$isHidden: true,
									$isTitleHidden: true,
									$isEditMode: false,
									$css: "s-job-view-info-field",
									$skin: "s-job-view-progress-data",
									$isTopLabelAlignment: true
								}]
							}, {
								"$layoutType": "stack",
								"$autoSize": true,
								$css: $skin + "-actions",
								$items: [{
									$bind: "$abort",
									$isTitleHidden: true,
									$category: "link",
									$css: $skin + "-actions-link"
								}, {
									$bind: "$suspend",
									$isTitleHidden: true,
									$category: "link",
									$css: $skin + "-actions-link"
								}, {
									$bind: "$resume",
									$isTitleHidden: true,
									$category: "link",
									$css: $skin + "-actions-link"
								}, {
									$bind: "$cancel",
									$isTitleHidden: true,
									$category: "link",
									$css: $skin + "-actions-link"
								}, {
									$bind: "$download",
									$isTitleHidden: true,
									$category: "link",
									$css: $skin + "-actions-link"
								}]
							}, {
								$bind: "$clearJob",
								$isTitleHidden: true,
								$category: "link",
								$noText: true,
								$skin: "s-job-view-remove-btn"
							}]
						}, {
							$layoutType: "row",
							$css: "s-job-view-phase-info",
							$items: [{
								$layoutType: 'stack',
								$items: [{
									$bind: "jobTitle",
									$isHidden: true,
									$isEditMode: false,
									$css: "s-job-view-info-field"
								}, {
									$bind: "phase",
									$isHidden: true,
									$isEditMode: false,
									$css: "s-job-view-info-field"
								}, {
									$bind: "phaseDetail",
									$isHidden: true,
									$isEditMode: false,
									$css: "s-job-view-info-field"
								}]
							}, {
								$layoutType: 'stack',
								$css: "s-job-view-seconds",
								$items: [{
									$bind: "elapsedSeconds",
									$isHidden: true,
									$isEditMode: false,
									$css: "s-job-view-info-field"
								}, {
									$bind: "remainingSeconds",
									$isHidden: true,
									$isEditMode: false,
									$css: "s-job-view-info-field"

								}]
							}]
						}, {
							$isBoxCollapsable: true,
							$title: localize.jobs_diag,
							$isHidden: true,
							$isDiagBox: true,
							$layout: {
								$layoutType: "stack",
								$items: [{
									$isEditMode: false,
									$isTitleRowHidden: true,
									$isTitleHidden: true,
									$isPagerHidden: true,
									$isQuickDesignerEnabled: false,
									$isMenuRecordHidden: true,
									$bind: "jobDiagnoses"
								}]
							}
						}]
					}
				}
			}]
		}
	};
};
