// syracuse-ui/lib/jobs/setting/jobsViewerPrototype.js
"use strict";
exports.$prototype = {
    $properties: {
        $jobType: {
            $type: "application/x-choice",
            $title: "List of",
            $value: {
                $type: "application/x-integer",
                $enum: []
            },
            $capabilities: "sort,filter"
        },
        $jobDetails: {
            "$type": "application/x-array",
            $item: {
                $properties: {
                    jobTitle: {
                        $title: "Job title",
                        $type: "application/x-string",
                        $capabilities: "sort,filter,alphaTab"
                    },
                    phase: {
                        $type: "application/x-string",
                        $title: "Phase",
                        $capabilities: "sort,filter,alphaTab"
                    },
                    phaseDetail: {
                        $type: "application/x-string",
                        $title: "Phase detail",
                        $capabilities: "sort,filter,alphaTab"
                    },
                    progress: {
                        $title: "Progress",
                        $type: "application/x-quantity",
                        $value: {
                            $title: "{$value} {$unit}",
                            $type: "application/x-decimal",
                            $format: "0.00"
                        },
                        $unit: "percentage",
                        $capabilities: "sort,filter,alphaTab"
                    },
                    percentage: {
                        "$title": "Percentage",
                        "$type": "application/x-choice",
                        "$value": {
                            "$type": "application/x-integer",
                            "$enum": [{
                                "$value": 1,
                                "$title": "%"
                            }]
                        },
                        $capabilities: "sort,filter,alphaTab"
                    },
                    elapsedSeconds: {
                        $type: "application/x-integer",
                        $title: "Elapsed seconds",
                        $capabilities: "sort,filter,alphaTab"
                    },
                    remainingSeconds: {
                        $type: "application/x-integer",
                        $title: "Remaining seconds",
                        $capabilities: "sort,filter,alphaTab"
                    },
                    pollingMillis: {
                        $type: "application/x-integer",
                        $title: "Polling millis",
                        $capabilities: "sort,filter,alphaTab"
                    },
                    progressIcon: {
                        $type: "application/x-string",
                        $title: "Progress icon"
                    },
                    diagnoseError: {
                        $type: "application/x-string",
                        $title: "diagnose error"
                    },
                    diagnoseFatal: {
                        $type: "application/x-string",
                        $title: "diagnose fatal"
                    },
                    diagnoseWarning: {
                        $type: "application/x-string",
                        $title: "diagnose warning"
                    },
                    diagnoseInfo: {
                        $type: "application/x-string",
                        $title: "diagnose info"
                    },
                    diagnoseSuccess: {
                        $type: "application/x-string",
                        $title: "diagnose success"
                    }
                },
                $links: {
                    $clearJob: {
                        $title: "Remove from the list"
                    }
                }
            }
        }
    },
    $links: {
        $clearList: {
            $title: "Clear this list"
        }
    }
};
