// syracuse-ui/lib/jobs/setting/jobsViewerArticle.js
"use strict";
var $skin = "s-jobs-viewer";
exports.$item = {
    $category: "section",
    $skin: $skin + "-h1",
    $isTitleHidden: true,
    $layout: {
        $items: [{
            $category: "section",
            $isTitleHidden: true,
            $layout: {
                $layoutType: "row",
                $autoSize: true,
                $items: [{
                    $css: $skin + "-types-list",
                    $category: "field",
                    $bind: "$jobType",
                    $isEditMode: true,
                    $format: "$combo",
                    $isTopLabelAlignment: false
                }, {
                    $category: "link",
                    $bind: "$clearList",
                    $css: $skin + "-clear-list"
                }]
            }
        }, {
            $title: "Jobs",
            $skin: $skin + "-head",
            $css: $skin + "-head",
            $format: "cards",
            $bind: "$jobDetails",
            $isTitleHidden: true,
            $isPagerHidden: true,
            $isQuickDesignerEnabled: false,
            $isMenuRecordHidden: true,
            $cards: {
                $category: "section",
                $layout: {
                    $items: [{
                        $category: "section",
                        $isTitleHidden: true,
                        $layout: {
                            $layoutType: "row",
                            $autoSize: true,
                            $css: $skin + "-card",
                            $items: [{
                                $layoutType: "stack",
                                $css: "s-job-view-types-icons",
                                $skin: "s-job-view-types-icons",
                                $items: [{
                                    $bind: "$jobPrintIcon",
                                    $isTitleHidden: true,
                                    $category: "link",
                                    $noText: true,
                                    $skin: "s-job-view-type-print"
                                }, {
                                    $bind: "$jobOperationIcon",
                                    $isTitleHidden: true,
                                    $category: "link",
                                    $noText: true,
                                    $skin: "s-job-view-type-operation"
                                }]
                            }, {
                                $isTitleHidden: true,
                                $layout: {
                                    $layoutType: "stack",
                                    $items: [{
                                        $bind: "jobTitle",
                                        $isTitleHidden: true,
                                        $isHidden: true,
                                        $isEditMode: false
                                    }, {
                                        $bind: "phase",
                                        $isTitleHidden: true,
                                        $isHidden: true,
                                        $isEditMode: false
                                    
                                    }, {
                                        $bind: "phaseDetail",
                                        $isTitleHidden: true,
                                        $isHidden: true,
                                        $isEditMode: false
                                    }]
                                }
                            }, {
                                $isTitleHidden: true,
                                $layout: {
                                    $layoutType: "stack",
                                    $items: [{
                                        $bind: "$progressOn",
                                        $isTitleHidden: true,
                                        $category: "link",
                                        $noText: true,
                                        $skin: "s-job-view-progress-value"
                                    }, {
                                        $bind: "$progressOff",
                                        $isTitleHidden: true,
                                        $category: "link",
                                        $noText: true,
                                        $skin: "s-job-view-progress-value-dis"
                                    }, {
                                        $bind: "progress",
                                        $isHidden: true,
                                        $isEditMode: false
                                    }, {
                                        $bind: "elapsedSeconds",
                                        $isHidden: true,
                                        $isEditMode: false
                                    }, {
                                        $bind: "remainingSeconds",
                                        $isHidden: true,
                                        $isEditMode: false
                                    }]
                                }
                            }, {
                                $isTitleHidden: true,
                                $layout: {
                                    "$layoutType": "stack",
                                    "$autoSize": true,
                                    $css: $skin + "-actions",
                                    $items: [{
                                        $bind: "$abort",
                                        $isTitleHidden: true,
                                        $category: "link",
                                        $css: $skin + "-actions-link"
                                    }, {
                                        $bind: "$suspend",
                                        $isTitleHidden: true,
                                        $category: "link",
                                        $css: $skin + "-actions-link"
                                    }, {
                                        $bind: "$resume",
                                        $isTitleHidden: true,
                                        $category: "link",
                                        $css: $skin + "-actions-link"
                                    }, {
                                        $bind: "$cancel",
                                        $isTitleHidden: true,
                                        $category: "link",
                                        $css: $skin + "-actions-link"
                                    }, {
                                        $bind: "$download",
                                        $isTitleHidden: true,
                                        $category: "link",
                                        $css: $skin + "-actions-link"
                                    }]
                                }
                            }, {
                                $bind: "$clearJob",
                                $isTitleHidden: true,
                                $category: "link",
                                $noText: true,
                                $skin: "s-job-view-remove-btn"
                            }]
                        }
                    }, {
                        $isBoxCollapsable: true,
                        $title: "additional information",
                        $isHidden: true,
                        $isDiagBox: true,
                        $layout: {
                            $layoutType: "stack",
                            $items: [{
                                $bind: "diagnoseFatal",
                                $isTitleHidden: true,
                                $isHidden: true,
                                $isEditMode: false
                            }, {
                                $bind: "diagnoseError",
                                $isTitleHidden: true,
                                $isHidden: true,
                                $rows: true,
                                $isEditMode: false
                            }, {
                                $bind: "diagnoseWarning",
                                $isTitleHidden: true,
                                $isHidden: true,
                                $rows: true,
                                $isEditMode: false
                            }, {
                                $bind: "diagnoseInfo",
                                $isTitleHidden: true,
                                $isHidden: true,
                                $rows: true,
                                $isEditMode: false
                            }, {
                                $bind: "diagnoseSuccess",
                                $isTitleHidden: true,
                                $isHidden: true,
                                $rows: true,
                                $isEditMode: false
                            }]
                        }
                    }]
                }
            }
        }]
    }
};
