// syracuse-ui/lib/tests/array/multi/cardsArrayTest.js
"use strict";
var helpers = require("syracuse-core/lib/helpers");
var MultiArrayTest = require("syracuse-ui/lib/tests/array/multi/multiArrayTest").MultiArrayTest;

function CardsArrayTest() {}



exports.CardsArrayTest = helpers.defineClass(CardsArrayTest, MultiArrayTest, {
	getCardDefinition: function(isTabs) {
		return {
			$items: [{
				$category: "section",
				$layout: {
					$items: [{
						$layoutType: "row",
						$items: [{
							$bind: "xstring"
						}, {
							$bind: "xchoice"
						}]
					}, {
						$layoutType: isTabs ? "tabs" : "row",
						$items: [{
							$category: "section",
							$title: "tab1",
							$layout: {
								$items: [{
									$bind: "xdate"
								}]
							}
						}, {
							$category: "section",
							$title: "tab2",
							$layout: {
								$items: [{
									$bind: "xboolean"
								}, {
									$bind: "xdecimal"
								}]
							}
						}]
					}]
				}
			}]
		};
	},
	_defineArticle: function($item) {
		MultiArrayTest.prototype._defineArticle.call(this, {
			$items: [{
				$category: "section",
				$layout: {
					"$layoutType": "stack",
					$items: [{
						$bind: "xarray",
						$format: "cards",
						$isEditMode: true,
						$layout: this.getCardDefinition()
					}, {
						$bind: "xarray",
						$format: "cards",
						$layout: this.getCardDefinition(true)
					}, {
						$bind: "xarray",
						$format: "cards",
						$isEditMode: true,
						"$isMenuRecordHidden": true,
						$layout: {
							$items: [{
								$category: "section",
								$layout: {
									$items: [{
										$layoutType: "row",
										$items: [{
											$bind: "xstring"
										}, {
											$bind: "xchoice"
										}]
									}, {
										$layoutType: "row",
										$items: [{
											$category: "section",
											$title: "tab1",
											$layout: {
												$items: [{
													$category: "menus",
													$skin: "s-record-actions",
													$layout: {
														$layoutType: "row",
														"$autoSize": true,
														$items: [{
															$noText: true,
															$bind: "$create",
															$icon: {
																$mode: "iconText",
																$path: "list/s-list-popup-"
															}
														}, {
															$noText: true,
															$bind: "$edit",
															$icon: {
																$mode: "iconText",
																$path: "list/s-list-popup-"
															}
														}, {
															$noText: true,
															$bind: "$delete",
															$icon: {
																$mode: "iconText",
																$path: "list/s-list-popup-"
															}
														}, {
															"$isMenusBag": true,
															$category: "menus",
															$title: "-",
															"$isTitlePicker": true,
															"$isBoxCollapsable": true,
															"$isPopupContent": true,
															$skin: "s-mn-popus",
															$itemSkin: "s-mn-link"
														}]
													}
												}, {
													$bind: "xdate"
												}]
											}
										}, {
											$category: "section",
											$title: "tab2",
											$layout: {
												$items: [{
													$bind: "xboolean"
												}, {
													$bind: "xdecimal"
												}]
											}
										}]
									}]
								}
							}]
						}
					}]
				}
			}]
		});
	}
});
