// syracuse-ui/lib/bookmark/navigator.js
"use strict";


exports.navigator = {
	addClick: function() {
		var menuItem = this.parent;
		if (!menuItem.isBookmark) {
			if (!_view.find(menuItem.urlSeg)) {
				menuItem.isBookmark = true;
				_view.dataset.items.push({
					menuItem: {
						$url: menuItem.urlSeg.$url,
						description: menuItem.codeMenu.$description || "",
						title: menuItem.codeMenu.$title || ""
					}
				});
			}
		} else {
			menuItem.isBookmark = false;
			var mn = _view.find(menuItem.urlSeg);
			mn && _view.deleteMenu(mn);
		}
		_toggleButton(menuItem.bookmarkBtn, menuItem.isBookmark);
		_view.commit();
	},
	refresh: function() {
		var page = syra_site.mainPage;
		if (page && page.isNavigationPage && !page.isAdminMode) {
			var modules = page.boundFields && page.boundFields.modules;
			if (modules) {
				for (var ii = 0, jj = modules.length; ii < jj; ii++) {
					var module = modules[ii];
					for (var mm = 0, kk = module.records.length; mm < kk; mm++) {
						var subModules = module.records[mm].subModulesItem;
						if (subModules) {
							for (var pp = 0, nn = subModules.records.length; pp < nn; pp++) {
								this._refreshItems(subModules.records[pp].itemsField);
							}
						}
					}
				}
			}
		}
	},
	checkMenu: function(menu) {
		menu.urlSeg = syra_url.parse(menu.codeMenu.$url);
		syra_url.replaceSegmentEndpoint(menu.urlSeg, syra_site.userProfile.dataset.selectedEndpoint, true);
		menu.isBookmark = _view.find(menu.urlSeg);
		_toggleButton(menu.bookmarkBtn, menu.isBookmark);
	},
	_refreshItems: function(items) {
		for (var ii = 0, jj = items.records.length; ii < jj; ii++) {
			var record = items.records[ii];
			if (record.singleField.$variantItemKey == "menuItem") {
				this.checkMenu(record.singleField.variantItem);
			} else {
				this._refreshItems(record.singleField.variantItem.record.itemsField);
			}
		}
	}
};
