// syracuse-ui/lib/soap/soapPage.js
"use strict";
var helpers = require("syracuse-core/lib/helpers")
var DesktopPage = require("syracuse-ui/lib/article/page").DesktopPage;
var soapController = require("syracuse-ui/lib/soap/soapController");

function SoapPage(){
}

exports.SoapPage = helpers.defineClass(SoapPage, DesktopPage, {
    loadBox: function(initData){
        this.$responseCount = 0;
        this.$item.$layout.$layoutType = "tabs";
        this.$item.$layout.$items[0].$title = "Request";
        DesktopPage.prototype.loadBox.call(this, initData);
    },
    onMenuClick: function(menuItem){
        if (menuItem.$bind == "$invoke") {
            var self = this;
            if (self.validateFields()) {
                var invokeMenu = self.$prototype.$links.$invoke;
                self.formatMenuUrl(menuItem, null, function($url){
                    menuItem.$url = $url;
                    document.controller.sendRequest(self, {
                        $location: menuItem,
                        data: self.sendBag,
                        method: menuItem.$method,
                        noDisplayErr: menuItem.noDisplayErr || false
                    }, function(data, response){
                        var newTab = document.itemFactory.create({
                            $title: "Response " + (++self.$responseCount),
                            $category: "section",
							$layout:{
								$items: [] 
							}
                        }, self);
                        self.layoutContent.appendNewItem({
                            load: true,
                            newItem: newTab
                        });
                        var soapQuery = document.controller.parseUrl(menuItem.$url);
                        var $itemPage = {
                            $$container: newTab.$$body,
                            $category: "vignette",
                            $urlParts: soapQuery.$urlParts,
                            $representation: response.data
                        };
                        $itemPage.$urlParts.$facet = "$response";
                        document.itemFactory.loadPage($itemPage);
                        newTab.openBox(true);
                        //soapController.openResponsePage(menuItem.$url, response.data);
                    }, function(err){
                    });
                });
            }
            
        }
    },
    dispose: function(){
        DesktopPage.prototype.dispose.call(this);
    }
});
