// syracuse-ui/lib/soap/soapPage.js
"use strict";
var helpers = require("syracuse-core/lib/helpers");
var DesktopPage = require("syracuse-ui/lib/page/desktopPage").DesktopPage;

function SoapPage() {}

exports.SoapPage = helpers.defineClass(SoapPage, DesktopPage, {
	loadBox: function(initData) {
		this.$responseCount = 0;
		this.$item.$layout.$layoutType = "tabs";
		this.$item.$layout.$items[0].$title = "Request";
		DesktopPage.prototype.loadBox.call(this, initData);
	},
	onMenuClick: function(menuItem) {
		if (menuItem.$bind == "$invoke") {
			var self = this;
			if (self.validateFields()) {
				var invokeMenu = self.$prototype.$links.$invoke;
				document.site.urlMaker.formatMenuUrl(self, menuItem, null, function($url) {
					menuItem.$url = $url;
					document.controller.sendRequest(self, {
						$location: menuItem,
						data: self.sendBag,
						method: menuItem.$method,
						noDisplayErr: menuItem.noDisplayErr || false
					}, function(data, response) {
						var newTab = self.layoutContent.createChildItem({
							$title: "Response " + (++self.$responseCount),
							$category: "section",
							$isCloseable: true,
							$layout: {
								$items: []
							}
						});
						var soapQuery = document.controller.parseUrl(menuItem.$url);
						var $itemPage = {
							layoutSlot: newTab.body,
							$category: "vignette",
							$urlParts: soapQuery.$urlParts,
							$representation: response.data
						};
						$itemPage.$urlParts.$facet = "$response";
						var initData = $itemPage.$representation.$prototype;
						var databody = initData.body;
						delete initData.body;
						var page = document.site.loadNewPage($itemPage);
						delete initData.$description;
						initData.body = databody;
						page.applyChange(initData);
						newTab.openBox(true);
					}, function(data, response) {
						document.site.showDiagnoses({
							$diagnoses: data && data.length > 0 ? (data.indexOf("$diagnoses") >= 0 ? JSON.parse(data).$diagnoses : [{
								$message: data ? data : document.site.localize.http_empty_error
							}]) : [{
								$message: document.site.localize.default_err_msg
							}]
						});
					});
					// TODO: onerror function
				});
			}

		}
	}
});
