// syracuse-ui/lib/soap/soapController.js
"use strict";

exports.isSoapUrl = function($url) {
	return $url && ($url.indexOf("soap-") >= 0);
};
exports.openRequestPage = function(openerUrlSegments) {
	var soapQuery = syra_controller.parseUrl(openerUrlSegments.$url);
	soapQuery.sendRequest({}, function(data, response, requestUrl) {
		var $itemPage = {
			layoutSlot: document.createElement("div"),
			$isEditMode: true,
			$category: "soap",
			openerUrlSegments: soapQuery.openerUrlSegments,
			$representation: data
		};
		$itemPage.openerUrlSegments.$facet = "$request";
		//$itemPage.initData = $itemPage.initData || initData;
		$itemPage.openerUrlSegments = openerUrlSegments;
		$itemPage.openerUrlSegments.fullUrl = document.location.href;
		syra_site.onMainPageChange($itemPage);
	}, function(data, response) {
		syra_site.showDiagnoses({
			$diagnoses: data && data.length > 0 ? (data.indexOf("$diagnoses") >= 0 ? JSON.parse(data).$diagnoses : [{
				$message: data ? data : syra_site.localize.http_empty_error
			}]) : [{
				$message: syra_site.localize.default_err_msg
			}]
		});
	});
	// TODO: onerror function
};
