// syracuse-ui/lib/soap/soapController.js
"use strict";

exports.isSoapUrl = function($url) {
	return $url && ($url.indexOf("soap-") >= 0);
};
exports.openRequestPage = function(httpQuery) {
	var soapQuery = document.controller.parseUrl(httpQuery.$url);
	soapQuery.sendRequest({}, function(data, response, requestUrl) {
		var $itemPage = {
			layoutSlot: document.createElement("div"),
			$isEditMode: true,
			$category: "soap",
			$urlParts: soapQuery.$urlParts,
			$representation: data
		};
		$itemPage.$urlParts.$facet = "$request";
		//$itemPage.initData = $itemPage.initData || initData;
		$itemPage.httpQuery = httpQuery;
		$itemPage.httpQuery.fullUrl = document.location.href;
		document.site.onMainPageChange($itemPage);
	}, function(data, response) {
		document.site.showDiagnoses({
			$diagnoses: data && data.length > 0 ? (data.indexOf("$diagnoses") >= 0 ? JSON.parse(data).$diagnoses : [{
				$message: data ? data : document.site.localize.http_empty_error
			}]) : [{
				$message: document.site.localize.default_err_msg
			}]
		});
	});
	// TODO: onerror function
};
