// etna/lib/supervisor/builtins/ACTXCACHE.ts
import { _ } from 'streamline-runtime';
import * as util from '../../supervisor/util';
import { types as coreTypes }  from 'syracuse-core';
const Helper = util.Helper;
import * as constants from '../../supervisor/constants';
import * as variables from '../../engine/runtime/variables';
import * as bcd from '../../engine/runtime/bcd';
import { Supervisor } from '../../supervisor/supervisor';
import { Class } from '../../supervisor/meta/class';

export class ACTXCACHE extends Helper {
	supervisor: Supervisor;
	$exported: boolean;
	constructor(superv: Supervisor) {
		super();
		this.supervisor = superv;
		this.$exported = true;
	}
	init(_: _) {
		this.class = null;
		return this;
	}
	get(_: _,  name: string) {
		return this.supervisor.load<Class>(_, 'Class', name);
	}
};

export function getInstance(_: _,  clas: Class, key: string) {
	// TODO: should check that class is eligible for context caching;
	const cache = clas.cachedInstances || (clas.cachedInstances = {});
	var instance = clas.cachedInstances[key];
	if (instance === undefined) {
		instance = clas.createInstance(_);
		var status = instance.read(_, [key]);
		if (status === constants.CST_AERROR) instance = null;
		cache[key] = instance;
	}
	return instance;
}

function getProp(_: _,  clas: Class, key: string, prop: string, type: string) {
	const instance = getInstance(_, clas, key);
	return instance ? instance.get(_, prop) : variables.types[type].default;
};

export const cacheMethods = {
	AGETVALNUM(_: _,  key: string, prop: string): number {
		return getProp(_, this, key, prop, 'I');
	},
	AGETVALDAT(_: _,  key: string, prop: string): coreTypes.date.DateValue {
		return getProp(_, this, key, prop, 'D');
	},
	AGETVALDEC(_: _,  key: string, prop: string): bcd.BCD {
		return getProp(_, this, key, prop, 'N');
	},
	AGETVALCHAR(_: _,  key: string, prop: string): string {
		return getProp(_, this, key, prop, 'S');
	}
};
