// etna/lib/supervisor/builtins/ASYRRCV.ts
import { _ } from 'streamline-runtime';
import * as util from '../../supervisor/util';
import { Supervisor } from '../supervisor';
import { X3Error } from '../../engine/runtime/errors';

export class Collection {
	name: string;
	collection: boolean;
	values: any[];
	constructor(name: string) {
		this.name = name;
		this.collection = true;
		this.values = [];
	}
	add(value: any) {
		this.values.push(value);
	};
	index(_: _,  offset: number) {
		if (offset > this.values.length) throw new X3Error(8, "Index incorrect ARCV." + name + ":" + JSON.stringify(this.values));
		return this.values[offset - 1];
	};
	x3Maxtab() {
		return this.values.length;
	};
}

export class ASYRRCV extends util.Helper {
	supervisor: Supervisor;
	qs: any;
	PARAMVAL: Collection;
	PARAMCOD: Collection;
	$exported: boolean;
	constructor(superv: Supervisor, qs: any) {
		super();
		this.supervisor = superv;
		this.$exported = true;
		this.qs = qs;
		this.PARAMVAL = new Collection("PARAMVAL");
		this.PARAMCOD = new Collection("PARAMCOD");
		Object.keys(qs).forEach(key => {
			this.PARAMCOD.add(key);
			this.PARAMVAL.add(qs[key]);
		});
	}
	get(_: _,  name: string) {
		switch (name) {
			case "PARAMVAL":
				return this.PARAMVAL;
			case "PARAMCOD":
				return this.PARAMCOD;
		}
	}
	paramval(_: _,  paramcod: string) {
		for (var i = 1, max = this.PARAMCOD.x3Maxtab(); i <= max; i++) {
			if (this.PARAMCOD.index(_, i) === paramcod) return this.PARAMVAL.index(_, i);
		}
		return "";
	}
};
