// syracuse-tablet/html/js/helpers/native/test/nativeTestVoiceCommands.js
"use strict";

var notifications = require('syracuse-tablet/html/js/helpers/notifications');
var jsutils = require('syracuse-tablet/html/js/helpers/jsutils');


var _micCss = {
	"color": "#41a940",
	"float": "right",
	"padding-left": "8px",
	"font-size": ".7em",
	"top": "8px",
	"position": "relative"
};
var _phraseListHtml = '\
	<div id="phraseListId" title="Test voice commands">\
		<div class="dropdown">\
	  		<span class="dropdown-toggle fa fa-microphone" data-toggle="dropdown" aria-expanded="true">\
	  		</span>\
	  		<ul class="dropdown-menu dropdown-menu-left" role="menu">\
				{{#each list}}\
					<li><a style="" draggable="false" href="#">{{title}}</a></li>\
				{{/each}}\
			</ul>\
		</div>\
	</div>';

var _getPhraseListMenu = function(ctx) {
	return Handlebars.compile(_phraseListHtml)(ctx);
};



var _TestNativeWrapper = function() {


	this.notifHistoryPush = function() {
		this.updateTestMenu();
	};
	this.updateTestMenu = function() {
		$("#phraseListId").off().remove();
		if (!this.phraseList) return;
		$(_getPhraseListMenu({
			list: this.phraseList
		})).css(_micCss).prependTo($('.s-m-header-nav'));
		$("#phraseListId li").click(jsutils.bindFn(this.onClickPhrase, this));
	};

	this.onClickPhrase = function(event) {
		var $$t = $(event.target);
		var text = $$t.text();
		if (!text || text.trim().length === 0) return;
		setTimeout(function() {
			window.smVoiceJS.executeVoiceCommand({
				command: "openBookmark",
				text: text
			});
		});
	};

	// phraseList: Array[24]
	this.phraseList = null;

	this.setPhraseListSupported = function() {
		return true;
	};

	this.setPhraseList = function(data) {
		data = data ? JSON.parse(data) : null;
		this.phraseList = null;
		if (data && data.phraseList && data.phraseList.length > 0) {
			var self = this;
			this.phraseList = [];
			data.phraseList.forEach(function(item) {
				self.phraseList.push({
					title: item
				});
			});
		};
		this.updateTestMenu();
	};


	notifications.subscribe(this, ["sm.history.push"]);
};



exports.create = function() {
	var url = jsutils.getCurrentUrl();
	if (window.smVoiceJS && url.query && url.query.voiceCommand) {
		try {
			window.smVoiceJS.executeVoiceCommand({
				command: "openBookmark",
				text: url.query.voiceCommand
			});
		} catch (e) {}
	}
	return new _TestNativeWrapper();
};
