// syracuse-tablet/html/js/helpers/native/test/nativeTestVoiceCommands.js
"use strict";

var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("nativeVoice", true);
var notifications = require('syracuse-tablet/html/js/helpers/notifications');
var jsutils = require('syracuse-tablet/html/js/helpers/jsutils');

var _phraseListHtml = '\
	<div id="phraseListId" class="btn-group s-m-link">\
		<div class="dropup">\
	  		<button type="button" class="btn btn-primary dropdown-toggle fa fa-microphone" data-toggle="dropdown" aria-expanded="true">\
	  		</button>\
	  		<ul class="dropdown-menu dropdown-menu-left" role="menu">\
				{{#each list}}\
					<li><a style="" draggable="false" href="#">{{title}}</a></li>\
				{{/each}}\
			</ul>\
		</div>\
	</div>';

var _getPhraseListMenu = function(ctx) {
	return Handlebars.compile(_phraseListHtml)(ctx);
};



var _TestNativeWrapper = function() {

	this.notifWelcomeApplication = function() {
		$("#phraseListId").off().remove();
		if (!this.phraseList) return;
		$(_getPhraseListMenu({
			list: this.phraseList
		})).appendTo($('.s-m-footer-left'));
		$("#phraseListId li").click(jsutils.bindFn(this.onClickPhrase, this));
	};

	this.onClickPhrase = function(event) {
		var $$t = $(event.target);
		var text = $$t.text();
		if (!text || text.trim().length === 0) return;
		setTimeout(function() {
			window.smVoiceJS.executeVoiceCommand({
				command: "openBookmark",
				text: text
			});
		});
	};

	// phraseList: Array[24]
	this.phraseList = null;

	this.setPhraseList = function(data) {
		data = data ? JSON.parse(data) : null;
		this.phraseList = null;
		if (data.phraseList && data.phraseList.length > 0) {
			var self = this;
			this.phraseList = [];
			data.phraseList.forEach(function(item) {
				self.phraseList.push({
					title: item
				});
			});
		};
	};


	notifications.subscribe(this, ["sm.welcomeApplication"]);
};

exports.create = function() {
	return new _TestNativeWrapper();
};
