// syracuse-tablet/html/js/helpers/native/test/nativeTestHelpers.js
"use strict";
var globals = require('syracuse-tablet/html/js/helpers/globals');


var _handlebars = {
	mailTo: ' <a id ="testNativeMailto" href="mailto:?subject={{subject}}&body={{body}}" class="s-m-link">\
		<i class="fa fa-envelope">\
		</i>\
		<span>{{title}}</span>\
		</a>',
	handleLink: ' <a id ="testNativeHandleLink" href="{{href}}" class="s-m-link">\
		<i class="fa fa-hand-pointer-o">\
		</i>\
		<span>{{title}}</span>\
		</a>'
};
var _getHtml = function(tmpl, ctx) {
	return Handlebars.compile(_handlebars[tmpl])(ctx);
};
/**
 * Call mobile application interface
 */
exports.callAppNativeInterface = function(moduleName, actionName) {
	var app = globals.getApplication();
	if (!app) return null;
	return app.nativeWrapperJSInterface.apply(app, arguments);
};

/**
 * Add a clickable email icon
 * options.subject	subject
 * options.url		url
 * options.timeout	timeout
 */
exports.addTestEmail = function($$parent, options) {
	if (!options) return;
	$(_getHtml("mailTo", {
		title: "Send email",
		subject: encodeURIComponent(options.subject || "No subject"),
		body: encodeURIComponent(options.url)
	})).appendTo($$parent);
	setTimeout(function() {
		$("#testNativeMailto").off().remove();
	}, options.timeout || 5000);
};
/**
 * Add a clickable hand icon
 * options.clickHandler	callback on click
 * options.url			link url
 * options.timeout		timeout
 */
exports.addHandleLink = function($$parent, options) {
	if (!options) return null;
	var $$elmt = $(_getHtml("handleLink", {
		title: options.title || "HandleLink",
		href: options.url || "#"
	})).appendTo($$parent);
	if (options.css) {
		$$elmt.css(options.css);
	}
	if (options.clickHandler) {
		$$elmt.click(function() {
			options.clickHandler();
		});
	}
	setTimeout(function() {
		$("#testNativeHandleLink").off().remove();
	}, options.timeout || 5000);
	return $$elmt;
};

exports.notify = function(title, text) {
	globals.getModal().notify({
		severityClass: "success",
		title: title,
		body: text,
		onlyWebapp: true
	});
};
