// syracuse-tablet/html/js/helpers/native/test/nativeTestPenInput.js
"use strict";

var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("nativePenInput");
var globals = require('syracuse-tablet/html/js/helpers/globals');

var _TestNativeWrapper = function() {

	this.GetPenTextInputSupported = function() {
		return true;
	};
	this.GetPenTextInput = function() {
		setTimeout(function() {
			log && log("Pen Input GET ");
			window.smPenInputJS.penInputTextFinished({
				Action: "ok",
				Matches: ["Pencil input..."]
			});
		}, 100);
	};
	this.GetAnnotatedImageSupported = function(options) {
		return true;
	};
	this.GetAnnotatedImage = function(options) {
		var img = "";
		setTimeout(function() {
			var data = {
				Action: "ok",
				ImageData: "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAUCAIAAAAC64paAAAACXBIWXMAAAsSAAALEgHS3X78AAAAGklEQVR42mOYacxANmIY1TyqeVTzqOaB1QwApdw+0KHEWp0AAAAASUVORK5CYII="
			};
			log && log("Annotated Image GET ");
			window.smPenInputJS.penAnnotateImageFinished(data);
		}, 100);
	};

	this.GetSignatureImageSupported = function(options) {
		return true;
	};
	this.GetSignatureImage = function(options) {
		var img = "";
		setTimeout(function() {
			var data = {
				Action: "ok",
				ImageData: "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAUCAIAAAAC64paAAAACXBIWXMAAAsSAAALEgHS3X78AAAAGklEQVR42mOYacxANmIY1TyqeVTzqOaB1QwApdw+0KHEWp0AAAAASUVORK5CYII="
			};
			log && log("Signature Image GET ");
			window.smPenInputJS.penSignatureImageFinished(data);
		}, 100);
	};

};

exports.create = function() {
	return new _TestNativeWrapper();
};
