// syracuse-tablet/html/js/helpers/native/test/nativeTestShare.js
"use strict";

var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("nativeShare", true);

var _mailTo = ' <a id ="testNativeShareMailto" href="mailto:?subject={{subject}}\Test email&body={{body}}" class="s-m-link">\
					<i class="fa fa-envelope">\
					</i>\
					<span>{{title}}</span>\
					</a>';
var _getMailto = function(ctx) {
	return Handlebars.compile(_mailTo)(ctx);
};

var _TestNativeWrapper = function() {

	this.supportsShare = function() {
		return true;
	};

	this.sharePageLink = function(data) {
		data = data ? JSON.parse(data) : null;
		var $$parent = $('.s-m-footer .s-m-link[data-action="nativeShare"]').parent();
		$(_getMailto({
			title: "Send email",
			subject: encodeURIComponent(data.DeepLink.Title),
			body: encodeURIComponent(data.DeepLink.Url)
		})).appendTo($$parent);
		setTimeout(function() {
			$("#testNativeShareMailto").remove();
		}, 5000);
		if (!window.smShareJS) return;
		setTimeout(function() {
			window.smShareJS.shareFinished({
				action: "ok",
			});
		});
	};
};

exports.create = function() {
	return new _TestNativeWrapper();
};
