// syracuse-tablet/html/js/helpers/native/test/nativeTestShare.js
"use strict";

var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("nativeShare");
var globals = require('syracuse-tablet/html/js/helpers/globals');
var testHelpers = require('syracuse-tablet/html/js/helpers/native/test/nativeTestHelpers');

var _TestNativeWrapper = function() {

	this.supportsShare = function() {
		return true;
	};


	this.SharePageLinkSupported = function() {
		return true;
	};
	this.SharePageLink = function(data) {
		var parsedData = data ? JSON.parse(data) : null;
		if (!parsedData) return;
		// Add icons to test email and click on tile
		testHelpers.addTestEmail($('.s-m-footer-left'), {
			subject: parsedData.DeepLink.Title,
			url: parsedData.DeepLink.Url
		});
		testHelpers.addHandleLink($('.s-m-footer-left'), {
			clickHandler: function() {
				setTimeout(function() {
					globals.getApplication().gotoWelcomeApplication().then(function() {
						setTimeout(function() {
							window.smShareJS.handleLink(parsedData.DeepLink);
						});
					});
				});
			}
		});
		if (!window.smShareJS) return;
		setTimeout(function() {
			window.smShareJS.shareFinished({
				action: "ok",
			});
		});
	};
};

exports.create = function() {
	return new _TestNativeWrapper();
};
