// syracuse-tablet/html/js/helpers/native/test/nativeTestApp.js
"use strict";

var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("nativeApp");
var localStorage = require('syracuse-tablet/html/js/storage/localStorage');
var nativeHelpers = require('syracuse-tablet/html/js/helpers/native/nativeHelpers');

var _TestNativeWrapper = function() {

	this.nativeWrapperSupported = function() {
		return true;
	};
	this.openExternalLinkSupported = function() {
		return true;
	};

	// This will hide html back button and only allow the one of the browser toolbar
	this.hardwareBackButtonSupported = function() {
		return true;
	};

	/**
	 *
	 */
	this.webappLoaded = function() {};
	/**
	 * historyContext is s a string
	 */
	this.onWebappUnloaded = function() {
		log && log("webappUnloaded");
		if (!window.smWrapperJS) return;
		try {
			this.setReconnectContext(nativeHelpers.callAppNativeInterface("nativeApp", "applicationClosing"));
		} catch (e) {
			return e.message;
		}
	};
	/**
	 *
	 */
	this.webappConfigServer = function() {};
	this.webappSelectServer = function() {};
	/**
	 * context is s a string
	 */
	this.setReconnectContext = function(context) {
		try {
			log && log("setReconnectContext", context);
			localStorage.setItem("$reconnectContext", context);
		} catch (e) {
			log && log("setReconnectContext", JSON.stringify(e));
		}
	};
	/**
	 * Returns a String
	 */
	this.getReconnectContext = function() {
		try {
			var context = localStorage.getItem("$reconnectContext");
			log && log("getReconnectContext", context);
			return context;
		} catch (e) {
			log && log("getReconnectContext", JSON.stringify(e));
			return null;
		}
	};
	/**
	 * Returns a String
	 */
	this.getReconnectToken = function() {
		try {
			var token = localStorage.getItem("$reconnectToken");
			log && log("getReconnectToken", token);
			return token;
		} catch (e) {
			log && log("getReconnectToken", JSON.stringify(e));
			return null;
		}
	};
	/**
	 * token is a string
	 *{
	 *		SyracuseSidLogin: token,
	 *		UserLogin: userLogin
	 *};
	 */
	this.setReconnectToken = function(token, userLogin) {
		try {
			log && log("setReconnectToken", token, userLogin);
			token = token ? JSON.parse(token) : null;
			token = token && token.SyracuseSidLogin;
			localStorage.setItem("$reconnectToken", token);
		} catch (e) {
			log && log("setReconnectToken", JSON.stringify(e));
		}
	};
};
exports.create = function() {
	return new _TestNativeWrapper();
};
