// syracuse-tablet/html/js/helpers/native/test/nativeTestX3WUPApp.js
"use strict";

var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("X3WUPApp");
var localStorage = require('syracuse-tablet/html/js/storage/localStorage');

var _TestNativeWrapper = function() {
	/**
	 *
	 */
	this.onWebappLoaded = function() {};
	/**
	 * historyContext is s a string
	 */
	this.onWebappUnloaded = function() {
		if (!window.smWrapperJS) return;
		try {
			this.setReconnectContext(window.smWrapperJS.applicationClosing());
		} catch (e) {
			console.log(e);
			return e.message;
		}
	};
	/**
	 *
	 */
	this.onWebappConfigServer = function() {};
	/**
	 * context is s a string
	 */
	this.setReconnectContext = function(context) {
		try {
			null && log && log("setReconnectContext", context);
			localStorage.setItem("$reconnectContext", context);
		} catch (e) {
			console.log(e);
		}
	};
	/**
	 * Returns a String
	 */
	this.getReconnectContext = function() {
		try {
			var context = localStorage.getItem("$reconnectContext");
			null && log && log("getReconnectContext", context);
			return context;
		} catch (e) {
			console.log(e);
			return null;
		}
	};
	/**
	 * Returns a String
	 */
	this.getReconnectToken = function() {
		try {
			var token = localStorage.getItem("$reconnectToken");
			log && log("getReconnectToken", token);
			return token;
		} catch (e) {
			console.log(e);
			return null;
		}
	};
	/**
	 * token is s a string
	 */
	this.setReconnectToken = function(token) {
		try {
			log && log("setReconnectToken", token);
			localStorage.setItem("$reconnectToken", token);
		} catch (e) {
			console.log(e);
		}
	};
};
exports.create = function() {
	return new _TestNativeWrapper();
};
