// syracuse-tablet/html/js/helpers/native/test/nativeTestShare.js
"use strict";

var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("nativeShare", true);
var globals = require('syracuse-tablet/html/js/helpers/globals');

var _mailTo = ' <a id ="testNativeShareMailto" href="mailto:?subject={{subject}}&body={{body}}" class="s-m-link">\
					<i class="fa fa-envelope">\
					</i>\
					<span>{{title}}</span>\
					</a>';

var _handleLink = ' <a id ="testNativeShareHandleLink" href="#" class="s-m-link">\
					<i class="fa fa-hand-pointer-o">\
					</i>\
					<span>{{title}}</span>\
					</a>';
var _getMailto = function(ctx) {
	return Handlebars.compile(_mailTo)(ctx);
};
var _getHandleLink = function(ctx) {
	return Handlebars.compile(_handleLink)(ctx);
};


var _TestNativeWrapper = function() {

	this.supportsShare = function() {
		return true;
	};

	this.sharePageLink = function(data) {
		var parsedData = data ? JSON.parse(data) : null;
		if (!parsedData) return;
		// Add a 'Send mail' link in the footer and remove it after 5 seconds
		var $$parent = $('.s-m-footer-left');
		$(_getMailto({
			title: "Send email",
			subject: encodeURIComponent(parsedData.DeepLink.Title),
			body: encodeURIComponent(parsedData.DeepLink.Url)
		})).appendTo($$parent);
		$(_getHandleLink({
			title: "HandleLink"
		})).appendTo($$parent).click(function() {
			setTimeout(function() {
				globals.getApplication().gotoWelcomeApplication().then(function() {
					setTimeout(function() {
						window.smShareJS.handleLink(parsedData);
					});
				});
			});
		});
		setTimeout(function() {
			$("#testNativeShareMailto").off().remove();
			$("#testNativeShareHandleLink").off().remove();
		}, 100000);
		if (!window.smShareJS) return;
		setTimeout(function() {
			window.smShareJS.shareFinished({
				action: "ok",
			});
		});
	};
};

exports.create = function() {
	return new _TestNativeWrapper();
};
