// syracuse-tablet/html/js/helpers/native/test/nativeTestLiveTile.js
"use strict";

var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("nativeLiveTile");
var globals = require('syracuse-tablet/html/js/helpers/globals');
var testHelpers = require('syracuse-tablet/html/js/helpers/native/test/nativeTestHelpers');

var _TestNativeWrapper = function() {
	/**
	 * OPTIONS
		{
		  "Id": "x3.erp.GX3APP.EXPENENTM.$query_http://pc101329.sagefr.adinternal.com:8124/mobile1/x3/erp/GX3APP/EXPENSES?representation=EXPENENTM.$query",
		  "Title": "Frais non visés",
		  "Description": "TEST",
		  "BackgroundColor": "#ff5800",
		  "ForegroundColor": "#41a940",
		  "SdataUrl": "http://pc101329.sagefr.adinternal.com:8124/mobile1/x3/erp/GX3APP/EXPENSES?representation=EXPENENTM.$query",
		  "DeepLink": {
		    "Title": "Frais non visés",
		    "Url": "http://pc101329.sagefr.adinternal.com:8124/syracuse-tablet/html/open_link.html?testNativeWrapper=true&context=1a8be16b-883a-45be-9b2e-1c96c29226fe"
		  },
		  "BindArray": "$resources",
		  "BindDetail": ""
		}
	 */
	this.AddLiveTileSupported = function() {
		return true;
	};
	this.AddLiveTile = function(options) {
		if (!options) return;
		try {
			options = JSON.parse(options);
		} catch (e) {
			return;
		}
		console.log(JSON.stringify(options, null, 2));
		var id = options.Id;
		var title = [];
		if (options.Title) {
			title.push(options.Title);
		}
		if (options.Description) {
			title.push(options.Description);
		}
		title = title.join(" - ");
		setTimeout(function() {
			log && log("LIVE TILE ADD - Succeeded - " + title);
			testHelpers.addHandleLink($('.navbar.s-m-side-panel'), {
				url: options.DeepLink.Url,
				timeout: 30000,
				title: title,
				css: {
					padding: "10px",
					color: options.ForegroundColor,
					backgroundColor: options.BackgroundColor
				}
			});
			window.smLiveTileJS.addLiveTileFinished(null);
		}, 100);
	};
};

exports.create = function() {
	return new _TestNativeWrapper();
};
