// syracuse-ui/lib/field/filter/filterField.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');


function _getOperators($type) {
	switch ($type) {
		case "application/x-date":
		case "application/x-time":
		case "application/x-datetime":
		case "application/x-decimal":
		case "application/x-real":
		case "application/x-integer":
			return ["none", "gt", "ge", "lt", "le", "eq", "ne", "between"];
			break;
		case "application/x-string":
		case "application/x-reference":
			return ["none", "gt", "ge", "lt", "le", "eq", "ne", "like_s", "like", "between"];
			break;
		default:
			return ["none", "eq", "ne"];
			break;
	}
}

var _popupItem, _popup;

function FilterField() {}

exports.FilterField = helpers.defineClass(FilterField, null, {
	createField: function($property, $item, filterBlock) {
		this.filterBlock = filterBlock;
		syra_store.add(this);
		this.$field = this.$item = $item;
		this.$bind = $property;
		this.$skin = filterBlock.$skin;

		this.layoutSlot = document.createElement("div");
		this.layoutSlot.className = this.$skin + "-field";
		this.layoutSlot.syraItem = this.id;
		this.filterBlock.filterBlockContainer.appendChild(this.layoutSlot);
		this.labelAndMenu = document.createElement("div");
		this.labelAndMenu.className = this.$skin + "-field-header";

		this.layoutSlot.appendChild(this.labelAndMenu);
		this.valuesSlot = document.createElement("div");
		this.valuesSlot.className = this.$skin + "-field-values";
		this.layoutSlot.appendChild(this.valuesSlot);


		this.addLabel();
		this.addValueFields();
		this.addMenu();
	},
	addValueFields: function() {
		var operators = _getOperators(this.$item.$type);
		this.fieldLow = this.addValueField(this.$bind + "$low");
		this.$$fieldLow = $(this.fieldLow);
		if (operators.indexOf("between") > -1) {
			this.fieldHigh = this.addValueField(this.$bind + "$high");
			this.$$fieldHigh = $(this.fieldHigh);
		}
		this.valuesSlot.className = this.$skin + "-field-values";
	},
	addValueField: function($bind) {
		var container = document.createElement("div");
		container.className = this.$skin + "-field";
		var field = this.filterBlock.articleParent.page.loadNewItem(this.valuesSlot.appendChild(container), {
			$bind: $bind,
			$isTopLabelAlignment: false,
			$isCellChild: true,
			$isMenusHidden: true,
			$isFilterMode: true,
			$inplace: true,
			$field: this.$item,
			$skin: "s-inplace"
		}, this.filterBlock);
		return field;
	},
	addLabel: function() {
		this.label = document.createElement("label");
		this.label.className = this.$skin + "-field-title";
		this.label.textContent = syra_site.expressionMaker.parse(this.filterBlock, this.$item.$title);
		this.labelAndMenu.appendChild(this.label);
	},
	addMenu: function() {
		this._operator = syra_menus.addIconButton(syra_local.flfilter_choiceOperator, this.$skin + "-field-menu", "onFilterOperatorClick", null, "filter_none");
		this.labelAndMenu.appendChild(this._operator);
	},
	onFilterOperatorClick: function(event, btn) {
		var self = this;
		if (_popup && _popupItem && _popupItem.syraItem == self.id) {
			_popup.close();
		} else {
			if (!_popupItem) {
				_popupItem = document.createElement("div");
				_popupItem.className = "s-filter-popup";
			}

			syra_site.dom.empty(_popupItem);
			_popupItem.syraItem = self.id;
			var operators = _getOperators(self.$item.$type);
			for (var ii = 0, jj = operators.length; ii < jj; ii++) {
				var op = operators[ii];
				var btn = syra_menus.addTextButton(syra_local["flFilter_" + op], "s-filter-popup-btn", "onFilterItemClick", null, "filter_" + op);
				btn.syraOperator = op;
				if (self.$filterOperator == op) {
					btn.className += " s-selected";
				}
				_popupItem.appendChild(btn);
			}
			_popup = syra_site.dialogManager.openPopup(self.filterBlock.articleParent, {
				content: self.filterBlock,
				slot: _popupItem,
				position: {
					my: "left top",
					at: "left bottom",
					of: $(self.layoutSlot)
				},
				onClose: function() {
					syra_site.dom.removeChild(_popupItem);
					_popup = _popupItem = null;
				}
			});
		}
	},
	onFilterItemClick: function(event, picker) {
		this.setOperator(picker.syraOperator);
		this.filterBlock.onNotifyDataChange(this.fieldLow);
		_popup.close();
	},
	setOperator: function(operator) {
		syra_menus.updateButtonIcon(this._operator, "", "filter_" + (this.operator = operator));
		if (this.operator === "none") {
			if (this.fieldLow) {
				$(this.fieldLow._dataValue).toggle(false);
			}
			if (this.fieldHigh) {
				$(this.fieldHigh._dataValue).toggle(false);
			}
		} else
		if (this.operator === "between") {
			if (this.fieldLow) {
				$(this.fieldLow._dataValue).toggle(true);
			}
			if (this.fieldHigh) {
				$(this.fieldHigh._dataValue).toggle(true);
			}
		} else {
			if (this.fieldLow) {
				$(this.fieldLow._dataValue).toggle(true);
			}
			if (this.fieldHigh) {
				$(this.fieldHigh._dataValue).toggle(false);
			}
		}
	},
	_format: function(field, $type, value) {
		switch ($type) {
			case "application/x-reference":
				var value = field.getInputValue();
				return "'" + (value || "") + "'";
			case "application/x-choice":
				var fieldType = (field.$field && field.$field.$value && field.$field.$value.$type) || "application/x-string";
				return this._format(field, fieldType, value);
			case "application/x-boolean":
				return (value == field.statusValues.on) ? "true" : "false";
			case "application/x-integer":
			case "application/x-decimal":
				return value;
			case "application/x-date":
			case "application/x-datetime":
			case "application/x-time":
				return "@" + value + "@";
		}
		return "'" + value + "'";
	},
	_formatValue: function(field, $type, value) {
		var self = this;
		if (!value) {
			return value;
		}

		switch ($type) {
			case "application/x-date":
			case "application/x-datetime":
			case "application/x-time":
				return value.toString(field.localFormat);
			case "application/x-boolean":
				return (value === "true") || (value === true) ? field.statusValues.on : field.statusValues.off;
			case "application/x-reference":
				// temporary fix for reference field in filter block widget - FK
				if (field.currentValue) {
					field.currentValue.$value = value;
					return field.currentValue;
				} else {
					return {
						$value: value
					};
				}
		}
		return value;
	},
	_ensureFilterField: function(field) {
		var $filterCode;
		if (field.$field.$type == "application/x-reference") {
			var $filterField = field.$field.$item[field.$reference.$value.$itemProp];
			if ($filterField) {
				$filterField = this.filterBlock.$prototype.$properties[field.$reference.$value.$prop];
				if ($filterField) {
					$filterCode = field.$reference.$value.$prop;
				}
			}
			if (!$filterField) {
				$filterField = field.$field.$item.$properties[field.$reference.$value.$prop];
				$filterCode = this.$bind + "." + field.$reference.$value.$prop;
			}
		}
		return $filterCode ? $filterCode : this.$bind;
	},
	getValue: function() {
		var self = this;
		var field = self.fieldLow;
		var value = self.getFieldValue(self.fieldLow);
		var highValue = self.getFieldValue(self.fieldHigh);
		var operator = self.operator;
		var $item = self.$item;
		var $filterCode = self._ensureFilterField(field); // temporary fix for reference field in filter block widget - FK
		if (operator != "none" && value != null) {
			var op = operator;
			var right;
			switch (op) {
				case "like":
					right = "'%" + ("" + value).replace(/'/g, "''") + "%'";
					break;
				case "like_s":
					right = "'" + ("" + value).replace(/'/g, "''") + "%'";
					op = "like";
					break;
				case "between":
					if (value && !highValue) {
						right = self._format(field, $item.$type, value);
						op = "ge";
					} else {
						if (!value && highValue) {
							right = self._format(field, $item.$type, highValue);
							op = "le";
						} else {
							if (value && highValue) {
								right = self._format(field, $item.$type, value) + " and " + self._format(field, $item.$type, highValue);
							} else {
								return "";
							}
						}
					}
					break;
				default:
					right = self._format(field, $item.$type, value);
					break;
			}
			return "(" + $filterCode + " " + op + " " + right + ")";
		}
	},
	getFieldValue: function(field) {
		var self = this;
		var value;

		if (!field)
			return null;

		if (field.$isEditMode === true) {
			switch (self.$item.$type) {
				case "application/x-date":
				case "application/x-datetime":
				case "application/x-time":
					value = field.getInputValue() === "" ? null : field.getInputValue();
					break;
				case "application/x-reference":
					// temporary fix for reference field in filter block widget - FK
					value = field.getInputValue();
					break;
				default:
					value = field.getDataValue();
					break;
			}
		} else {
			value = field._dataValue.textContent;
		}

		return value;
	},
	setFilterValues: function(operator, value, highValue) {
		var self = this;

		if (operator === "like") {
			if (value.length > 1 && value.charAt(value.length - 1) === "%" && value.charAt(0) !== "%") {
				operator = "like_s";
			}
			value = value.replace(/%/g, "");
		}

		if (operator === "like_s" && value.charAt(value.length - 1) === "%") {
			value = value.replace(/%/g, "");
		}

		self.setOperator(operator);
		switch (operator) {
			case "none":
				self.setFilterValue(self.fieldLow, null);
				self.setFilterValue(self.fieldHigh, null);
				break;
			case "between":
				self.setFilterValue(self.fieldLow, value);
				self.setFilterValue(self.fieldHigh, highValue);
				break;
			default:
				self.setFilterValue(self.fieldLow, value);
				self.setFilterValue(self.fieldHigh, null);
				break;
		}
	},
	setFilterValue: function(field, value) {
		if (field) {
			value = this._formatValue(field, this.$item.$type, value);
			field.setDataValue(value);
		}
	},
	applyModifyState: function(mode) {
		return; // obsolete
		if (this.fieldLow) {
			this.fieldLow.toggleEditMode(mode);
		}
		if (this.fieldHigh) {
			this.fieldHigh.toggleEditMode(mode);
		}
		this._modifyState = mode;
	},
	dispose: function() {
		_popup && _popup.close();
		syra_store.remove(this);
		this.fieldHigh && this.fieldHigh.dispose();
		this.fieldLow && this.fieldLow.dispose();
		syra_site.disposeObject(this);
	}
});
