// syracuse-ui/lib/field/filter/gridFilterBuilder.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Article = require("syracuse-ui/lib/article/article").Article;
var _filterMaker = require('syracuse-ui/lib/field/filter/filterMaker');
var _popupItem, _popup;

exports.dispose = function() {
	_popup && _popup.close();
};

function _onFilterOperatorClick(field, event, picker) {
	var record = field.articleParent;
	var options = {
		operator: field.$filterOperator,
		list: record.list,
		event: event,
		type: "operatorclick",
		doEvent: function() {
			if (_popup && _popupItem && _popupItem.syraItem == field.id) {
				_popup.close();
			} else {
				syra_site.dialogManager.closePopups();
				_ensureFilterField(field);
				if (!_popupItem) {
					_popupItem = document.createElement("div");
					_popupItem.className = "s-filter-popup";
				}
				syra_site.dom.empty(_popupItem);
				_popupItem.syraItem = field.id;
				var operators = _filterMaker.getFieldOperators(null, field, field.$filterField);
				for (var ii = 0, jj = operators.length; ii < jj; ii++) {
					var op = operators[ii];
					var btn = syra_menus.addTextButton(syra_local["flFilter_" + op], "s-filter-popup-btn", "onFilterItemClick");
					btn.syraOperator = op;
					if (field.$filterOperator == op) {
						btn.className += " s-selected";
					}
					btn.style.backgroundImage = "url('" + syra_site.$item.$iconPath + "page/s-filter-" + op + ".png')";
					_popupItem.appendChild(btn);
				}
				_popup = syra_site.dialogManager.openPopup(record.boxParent, {
					content: record,
					slot: _popupItem,
					position: {
						my: "left top",
						at: "left bottom",
						of: $(field.layoutSlot)
					},
					onClose: function() {
						syra_site.dom.removeChild(_popupItem);
						_popup = _popupItem = null;
					}
				});
			}
		}
	};
	record.page.externalAdapter.onFilterEvent(options);
}

function _onFilterItemClick(field, event, picker) {
	_setOperator(field, picker.syraOperator);
	_popup.close();
	field.page.externalAdapter.onFilterEvent({
		operator: field.$filterOperator,
		list: field.articleParent.list,
		filter: field,
		event: event,
		type: "operatorchange",
		doEvent: function() {
			var record = field.articleParent;
			var value = record.dataset[field.$item.$bind];
			if (field.$filterOperator != "none") {
				if (value != null && value != "") {
					if (field.$filterOperator == "empty" || field.$filterOperator == "notempty") {
						field.setDataValue(null);
					}
					_buildFilter(record);
				} else {
					if (field.$filterOperator == "empty" || field.$filterOperator == "notempty") {
						_buildFilter(record);
					}
				}
			} else {
				if (value != null && value != "") {
					field.setDataValue(null);
					field.notifyFieldChange(null);
				} else {
					_buildFilter(record);
				}
			}
		}
	});
}

function _ensureFilterField(field) {
	if (field.$field.$type == "application/x-reference") {
		field.$filterField = field.$field.$item[field.$reference.$value.$itemProp];
		if (field.$filterField) {
			field.$filterField = field.articleParent.$prototype.$properties[field.$reference.$value.$prop];
			if (field.$filterField) {
				field.$filterCode = field.$reference.$value.$prop;
			}
		}
		if (!field.$filterField) {
			field.$filterField = field.$field.$item.$properties[field.$reference.$value.$prop];
			field.$filterCode = field.$item.$bind + "." + field.$reference.$value.$prop;
		}
	}
}

function _setOperator(field, operator) {
	field.filterOperatorBtn.style.backgroundImage = "url('" + syra_site.$item.$iconPath + "page/s-filter-" + (field.$filterOperator = operator) + ".png')";
}

function _buildFilter(record) {
	var values = [];
	var $binds = Object.keys(record.boundFields);
	for (var ii = 0, jj = $binds.length; ii < jj; ii++) {
		var field = record.boundFields[$binds[ii]][0];
		if (!field.$isDisabled) {
			var value = _getCellValue(field);
			value && values.push(value);
		}
	}
	var filter = values.join(" and ");
	if (record.fixed_where) {
		filter = record.fixed_where + (filter ? (" and " + filter) : "");
	}
	record.list.fetcher.trigger(record.list, {
		filter: filter
	});
}


function _loadField(record, slot, $field, $bind) {
	var field = record.page.loadNewItem(slot.appendChild(document.createElement("div")), {
		$bind: $bind,
		$isTopLabelAlignment: false,
		$isCellChild: true,
		$isFilterMode: true,
		$isMenusHiddeen: true,
		$inplace: true,
		$field: $field
	}, record);
	if (field.input) {
		field.input.className += " s-list-filter-input";
	}
	field.$filterCode = $bind;
	field.$filterOperator = "none";
	field.filterOperatorBtn = syra_menus.addIconButton(syra_local.flfilter_choiceOperator, "s-list-filter-cell-picker", "onFilterOperatorClick");
	field.filterOperatorBtn.removeAttribute("href");
	field.filterOperatorBtn.style.backgroundImage = "url('" + syra_site.$item.$iconPath + "page/s-filter-none.png')";
	field._dataValue.insertBefore(field.filterOperatorBtn, field._dataValue.firstChild);
	field.onFilterOperatorClick = function(event, btn) {
		_onFilterOperatorClick(this, event, btn);
	};
	field.onFilterItemClick = function(event, btn) {
		_onFilterItemClick(this, event, btn);
	};
	return field;
}

exports.onNotifyDataChange = function(record, field, value) {
	record.dataset[field.$item.$bind] = value;
	if (value == null || value == "") {
		field.$filterOperator !== "none" && _setOperator(field, "none");
	} else {
		if (field.$filterOperator === "none" && field.$field) {
			var defaultFilter = "eq";
			switch (field.$field.$type) {
				case "application/x-date":
				case "application/x-time":
				case "application/x-datetime":
					defaultFilter = "ge";
					break;
				case "application/x-string":
					defaultFilter = "like";
					break;
			}
			_setOperator(field, defaultFilter);
		}
	}
	_buildFilter(record);
	return false;
};

function _getSortedFilterRcdFields(filterRcd) {
	return (Object.keys(filterRcd.fieldCellsMap)).sort(function(a, b) {
		return parseInt(a.substr(2), 10) - parseInt(b.substr(2), 10);
	});
}

function _getUIOrdereddBinds(list, filterRcd) {
	var ret;
	return ((ret = list.$userPreferences) && (ret = ret.$columns) && ret.length > 0) ? ret : _getSortedFilterRcdFields(filterRcd);
}

exports.getQuickSelValues = function(list) {
	var res = [];
	var record = list.builder.filterRecord;
	if (record) {
		var $binds = _getSortedFilterRcdFields(record);
		for (var ii = 0, jj = $binds.length; ii < jj; ii++) {
			var field = record.fieldCellsMap[$binds[ii]];
			if (field.$field) {
				var defaultFilter, value = field.getInputValue();
				switch (field.$field.$type) {
					case "application/x-string":
						defaultFilter = (value !== "" && field.$filterOperator === "none") ? "like_s" : (value === "" && field.$filterOperator !== "none") ? "none" : undefined;
						break;
					default:
						defaultFilter = (value !== "" && field.$filterOperator === "none") ? "ge" : (value === "" && field.$filterOperator !== "none") ? "none" : undefined;
						break;
				}
				defaultFilter && _setOperator(field, defaultFilter);
			}
			res.push(_getCellValue(field) || "");
		}
	}
	return res;
};

function _getCellValue(field) {
	if (field.$filterOperator != "none") {
		_ensureFilterField(field);
		return _filterMaker.getFieldValue({
			field: field,
			$bind: field.$item.$bind,
			$type: field.$field.$type,
			operator: field.$filterOperator,
			$filterCode: field.$filterCode,
			$filterField: field.$filterField
		});
	}
}

exports.getNextQuickSelField = function(list, bind, nextSel) {
	var ret = null,
		next, prev;
	var record = list.builder.filterRecord;
	if (record.$item && record.$item.$layout && record.$item.$layout.$items) {
		if (nextSel === undefined) {
			ret = record.boundFields[bind];
		} else {
			var cells = _getUIOrdereddBinds(list, record);
			for (var ii = 0, jj = cells.length; ii < jj; ii++) {
				if (cells[ii] === bind) {
					if (ii === 0) {
						next = cells[ii + 1];
						prev = bind;
					} else
					if (ii === (cells.length - 1)) {
						next = cells[0];
						prev = cells[ii - 1];
					} else {
						next = cells[ii + 1];
						prev = cells[ii - 1];
					}
					break;
				}
			}
			ret = record.boundFields[nextSel ? next : prev];
		}
	}
	return ret;
};



function _addFieldCell(col) {
	var record = col.list.builder.filterRecord;
	var td = record.cellsMap[col.$item.$bind] = document.createElement("td");
	td.className = "s-list-filter-cell";
	record[col.table.$rowKey].appendChild(td);
	if (col.$isHidden) {
		td.style.display = "none";
	}
	col.cssFilter = null;
	if (col.$field.$hasFilterCapability) {
		var field = _loadField(record, td, col.$field, col.$item.$bind);
		record.fieldCellsMap[col.$item.$bind] = field;
		col.cssRecordCellFilter = " s-grid-col-filter";
		switch (col.$field.$type) {
			case "application/x-choice":
			case "application/x-date":
			case "application/x-time":
			case "application/x-reference":
				col.cssRecordCellFilter += " s-grid-col-filter-type-btn";
		}
		return field;
	}
};

function _setWhere(record, where, field) {
	var $filterBind = field.$item.$bind;
	var $isRefField = false;
	// in case of reference field, appropriate $bind value is field.$reference.$value.$prop
	if (field.$field.$type == 'application/x-reference') {
		$filterBind = field.$reference.$value.$prop;
		$isRefField = true;
	}
	var exp = _filterMaker.parseWhere(where, $filterBind);
	if (exp && $filterBind == exp.$bind) {
		_setOperator(field, exp.$operator);
		// update dataset and field data value
		if ($isRefField) {
			record.dataset[field.$item.$bind] = {};
			record.dataset[field.$item.$bind][$filterBind] = exp.$value;
		} else {
			record.dataset[field.$item.$bind] = exp.$value;
		}
		field.setDataValue(exp.$value);
	}
}

function FilterRecord() {}

helpers.defineClass(FilterRecord, Article, {
	loadFilter: function() {
		var params = this.page.openerUrlSegments && this.page.openerUrlSegments.params;
		if (params && params.where) {
			this.where = params.where;
			if (this.page.lookupField == this.list) {
				this.fixed_where = params.where;
			}
		}
		var columns = this.list.builder.allColumns;
		for (var ii = 0, jj = columns.length; ii < jj; ii++) {
			var col = columns[ii];
			if (col && col.$field) {
				var field = _addFieldCell(col);
				if (field) {
					if (this.where) {
						this.where && _setWhere(this, this.where, field);
						if (field.$filterOperator != "none") {
							if (this.fixed_where) {
								field.disable(true);
								syra_site.dom.disableItem(field.filterOperatorBtn, true);
							}
						}
					}
				}
			}
		}
	},
	loadRecord: function(list) {
		this.list = list;
		this.isFilterArticle = this.isRecordArticle = true;
		this.$facet = "$filter";
		this.$isEditMode = true;
		this.$prototype = list.$prototype.$item;
		this.$prototype.$localization = list.page.$prototype.$localization;
		list.page.initializeNewItem(this, {
			$layout: {
				$items: list.$item.$layout.$items
			}
		}, list);
		this.loadBox();
	},
	//set focus on first field
	setFocus: function() {
		var $bind = this.list.builder.getNextFocusableField();
		var field = $bind && this.fieldCellsMap[$bind];
		field && field.focus();
	},
	drawBox: function() {
		this.cellsMap = {};
		this.fieldCellsMap = {};
		this.domItem = this.dataRow = document.createElement("tr");
		var builder = this.list.builder;
		builder.freezeTable && builder.freezeTable.headTable.appendChild(this.freezeRow = document.createElement("tr"));
		builder.scrollTable.headTable.appendChild(this.dataRow);
	},
	onNotifyDataChange: function(field, value) {
		return exports.onNotifyDataChange(this, field, value);
	}
});

exports.addRecord = function(list) {
	var record = new FilterRecord();
	record.loadRecord(list);
	return record;
};
