// syracuse-perfmon/lib/monitor.js
"use strict";
/// !doc
/// 
/// # Simple performance monitoring API
/// 
/// This API allows you to gather performance data in your JS code. Typical usage is:
/// 
/// ``` javascript
/// var perfmon = require('syracuse-perfmon');
/// 
/// function myFunction(p1, p2) {
///   var timing = perfmon.start(module, "myFunction", p1);
///   // my function body
///   // ...
///   timing.end();
/// }
/// ```
/// 
/// The collected performance data can be retrieved with an HTTP request on http://localhost:8124/perfmon/session-data.
/// A link for it has been added in the main menu panel.
/// 
var helpers = require('syracuse-core/lib/helpers');
var glob = require('streamline/lib/globals');

var Timing = helpers.defineClass(function(mod, tag, details) {
	this.module = mod && mod.id;
	this.tag = "" + tag;
	this.details = details == null ? "" : details.toString();
	this.start = Date.now();
}, null, {
	end: function() {
		var cx = (glob.context || (glob.context = {}));
		var timings = (cx.timings || (cx.timings = []));
		timings.push({
			module: this.module,
			tag: this.tag,
			details: this.details,
			start: this.start,
			end: Date.now(),
		});
	}
})


/// ## API
/// 
/// `var perfmon = require('syracuse-perfmon');`
/// 
/// * `var timing = perfmon.start(module, tag, details);`  
///   start a monitoring operation.  
///   `module` is the current node.js module. This variable is set automatically by the require machinery.  
///   `tag` identifies the function or operation.  
///   `details` gives details about current invocation of the operation (typically a parameter value).
/// * `timing.end()`  
///   ends the timing and records the performance data.
exports.start = function(mod, tag, details) {
	return new Timing(mod, tag, details);
}

function hierarchize(timings, root) {
	root.children = [];
	var n;
	while ((n = timings.pop()) && n.start >= root.start) {
		root.children.push(hierarchize(timings, n));
	}
	if (n) timings.push(n);
	return root;
}

function cssClass(millis) {
	if (millis < 10) return "low";
	if (millis < 100) return "medium";
	if (millis < 1000) return "high";
	return "very-high";
}

function format(n) {
	var mod = n.module.substring(n.module.indexOf('node_modules') + 13);
	var millis = (n.end - n.start);
	var s = '<li><span class="' + cssClass(millis) + '"><b>' + millis + " ms</b></span> " + //
	'<span title="' + mod + '" class="tag">' + n.tag + '</span> ' + //
	'<span class="details">' + n.details + '</span>';
	s += n.children ? ('<ul>' + n.children.map(format).join('') + '</ul>') : ''
	s += '</li>';
	return s;
}

var head = '<html>' + '<head><title>Syracuse Prototypes</title>' + //
'<link href="/syracuse-perfmon/lib/styles.css" rel="stylesheet" type="text/css"/>' + //
'</head><body>';
var tail = '</body></html>';

/// * `var collection = perfmon.collect(reset)`  
///   returns the array of timing records that have been collected.  
///   if `reset` is true the recorded data is reset.
exports.collect = function(reset) {
	if (!glob.context) return null;
	var timings = glob.context.timings;
	if (reset) glob.context.timings = null;
	return timings;
}
/// * var html = perfmon.toHtml(collections);  
///   formats collected data as HTML.  
///   `collections` is an array of timing collections.
///   Each element of the `collections` array is an array containing timing entries.
exports.toHtml = function(collections) {
	if (!collections) return head + 'Sorry, no data available. <a href="/syracuse-main/html/main.html?url=%3Frepresentation%3Dhome.%24dashboard">Navigate first!</a>' + tail; 
	var children = collections.map(function(timings) {
		timings = timings.slice(0);
		return hierarchize(timings, timings.pop());
	});
	return head + '<ul>' + children.map(format).join('') + '</ul>' + tail;
}
