// syracuse-perfmon/lib/monitor.js
"use strict";
/// !doc
/// 
/// # Simple performance monitoring API
/// 
/// This API allows you to gather performance data in your JS code. Typical usage is:
/// 
/// ``` javascript
/// var perfmon = require('syracuse-perfmon');
/// 
/// function myFunction(p1, p2) {
///   var timing = perfmon.start(module, "myFunction", p1);
///   // my function body
///   // ...
///   timing.end();
/// }
/// 
/// The collected performance data can be retrieved with an HTTP request on 
/// http://localhost:8124/perfmon/session-data.
/// A link for it has been added in the main menu panel.
var helpers = require('syracuse-core/lib/helpers');
var glob = require('streamline/lib/globals');

var Timing = helpers.defineClass(function(mod, tag, details) {
	this.module = mod && mod.id;
	this.tag = "" + tag;
	this.details = "" + details;
	this.start = Date.now();
}, null, {
	end: function() {
		var cx = (glob.context || (glob.context = {}));
		var timings = (cx.timings || (cx.timings = []));
		timings.push({
			module: this.module,
			tag: this.tag,
			details: this.details,
			start: this.start,
			end: Date.now(),
		});
	}
})


/// ## API
/// 
/// `var perfmon = require('syracuse-perfmon');`
/// 
/// * `var timing = perfmon.start(module, tag, details);
///   start a monitoring operation.  
///   `module` is the current node.js module. This variable is set automatically by the require machinery.  
///   `tag` identifies the function or operation.  
///   `details` gives details about current invocation of the operation (typically a parameter value).
/// * `timing.end()`  
///   ends the timing and records the performance data.
exports.start = function(mod, tag, details) {
	return new Timing(mod, tag, details);
}

function hierarchize(timings, root) {
	root.children = [];
	var n;
	while ((n = timings.pop()) && n.start >= root.start) {
		root.children.push(hierarchize(timings, n));
	}
	if (n) timings.push(n);
	return root;
}

function format(n) {
	var mod = n.module.substring(n.module.indexOf('node_modules') + 13)
	var s = '<li><b>' + (n.end - n.start) + "ms</b> " + mod + " " + n.tag + " " + n.details;
	s += n.children ? ('<ul>' + n.children.map(format).join('') + '</ul>') : ''
	s += '</li>';
	return s;
}

var head = '<html>' + '<head><title>Syracuse Prototypes</title>' + '<link href="/syracuse-ui/themes/markdown.css" rel="stylesheet" type="text/css"/>' + '</head><body>';

var tail = '</body></html>';

exports.toHtml = function(requests) {
	var children = requests.map(function(timings) {
		timings = timings.slice(0);
		return hierarchize(timings, timings.pop());
	});
	return head + '<ul>' + children.map(format).join('') + '</ul>' + tail;
}
