// @sage/syracuse-react/src/navigation/breadcrumb/components/module.tsx
"use strict";
import * as React from 'react';
import { Link, } from '../../../link/link';
import { Submodules } from './submodule';
import { SitemapModule } from '../../sitemap/groups';
import { IProps } from './props';

interface IModulesProps extends IProps {
    modules: SitemapModule[]
}

export class Modules extends React.Component<IModulesProps, {}> {
    render() {
        return <ul className="s_breadcrumb_modules">
            {this.props.modules.map((module) =>
                <Module
                    key={module.$uuid}
                    module={module}
                    onMenuClick={this.props.onMenuClick}
                    onExpand={this.props.onExpand}
                    expandedList={this.props.expandedList}
                    path={this.props.path}/>)}
        </ul>;
    }
}

interface IModuleProps extends IProps {
    module: SitemapModule
}
class Module extends React.Component<IModuleProps, {}> {
    get expanded(): boolean {
        return this.props.expandedList[this.props.module.$uuid];
    }
    render() {
        let css = (this.props.path.module == this.props.module ? "s_selected " : "") + "s_breadcrumb_module_link";
        return <li className="s_breadcrumb_module">
            <Link
                className={css}
                title={this.props.module.title}
                icon={this.expanded ? "dir_down" : "next"}
                iconText="left"
                value={this.props.module}
                onClick={this.props.onExpand} />
            {this.expanded &&
                <div className="s_breadcrumb_subgroup">
                    <Submodules {...this.props} />
                </div>}
        </li>;
    }
}
