// syracuse-search/lib/representations/queryResult.js
"use strict"

exports.representation = {
    $entityName: "queryResult",
    $facets: {
        $query: {
            $layout: {
                $layoutType: "tabs",
                $items: [{
                    $category: "section",
                    $title: "Functions",
                    $layout: {
                        $layoutType: "columns",
                        $layoutSubType: "50,50",
                        $items: [{
                            $items: [{
                                $bind: "$resources",
                                $format: "cards",
                                $cards: {
                                    $layout: {
                                        $items: [{
                                            $bind: "description",
                                            $isTitleHidden: true
                                        }, {
                                            $bind: "type",
                                            $isTitleHidden: true
                                        }]
                                    }
                                }
                            }],
                        }]
                    }
                }, {
                    $category: "section",
                    $title: "Data",
                    $layout: {
                        $layoutType: "columns",
                        $layoutSubType: "50,50",
                        $items: [{
                            $items: [{
                                $bind: "$resources",
                                $format: "cards",
                                $cards: {
                                    $layout: {
                                        $items: [{
                                            $bind: "description",
                                            $isTitleHidden: true
                                        }, {
                                            $bind: "type",
                                            $isTitleHidden: true
                                        }]
                                    }
                                }
                            }],
                        }]
                    }
                }]
            }
        },
        $search: {
            $copy: "$query"
        }
    }
}
