// syracuse-search/lib/representations/queryResult.js
"use strict";

exports.representation = {
	$entityName: "queryResult",
	$facets: {
		$query: {
			$layout: {
				$layoutType: "tabs",
				$items: [{
					$category: "section",
					$title: "Functions",
					$layout: {
						$layoutType: "row",
						$widths: "50,50",
						$items: [{
							$items: [{
								$bind: "$resources",
								$skin: "toto",
								$format: "cards",
								$layout: {
									$items: [{
										$bind: "description",
										$isTitleHidden: true,
										$css: "tutu"
									}, {
										$bind: "type",
										$isTitleHidden: true,
										$skin: "popi"
									}]
								}
							}],
						}]
					}
				}, {
					$category: "section",
					$title: "Data",
					$layout: {
						$layoutType: "row",
						$widths: "50,50",
						$items: [{
							$items: [{
								$bind: "$resources",
								$format: "cards",
								$layout: {
									$items: [{
										$bind: "description",
										$isTitleHidden: true
									}, {
										$bind: "type",
										$isTitleHidden: true
									}]
								}
							}],
						}]
					}
				}]
			}
		},
		$search: {
			$copy: "$query"
		}
	}
};
