// syracuse-ui/lib/mobile/field/vignette/htmlVignetteBuilder.js
"use strict";

exports.getSettings= function(field){
	
};
exports.load = function(field){
    var url = "http://www.lemonde.fr/" || field.$field.$location.$url;
	var url = field.$field.$location.$url;
    var title = field.$$title.text();
    var link = document.createElement("a");
    link.className = field.$skin + "-title-link";
    link.setAttribute("href", url);
    link.setAttribute("target", "_blank");
    link.setAttribute("title", field.$field.$location.$title);
    $(link).appendTo(field.$$title.empty()).text(title);
    var iframe = document.createElement("iframe");
    iframe.setAttribute("src", url);
    iframe.style.width = field.$item.$width || "100%";
    iframe.style.height = field.$item.$height || "100%";
    field.$$body.empty().append(field.$$iframe = $(iframe));
};
exports.maximize = function(field){
    exports.onWindowResize(field);
    document.site.addResizeListener( field);
};
exports.minimize = function(field){
    field.$$iframe[0].style.height = field.$item.$height || "100%";
    document.site.removeResizeListener(field);
};
exports.onWindowResize = function(field){
    var body = document.site.$$body;
    var maxHeight = document.site.$$body.height();
    var iframe = field.$$iframe[0];
    iframe.style.height = maxHeight + "px";
    iframe.style.height = (maxHeight - (body[0].scrollHeight - maxHeight)) + "px";
};
exports.getAuthoringWidget = function(field){
    return {
        $prototype: {
            "$properties": {
                $locationTitle: {
                    $type: "application/x-string",
                    $title: "Title"
                },
                $locationUrl: {
                    $type: "application/x-string",
                    $title: "url"
                },
                $width: {
                    $type: "application/x-string",
                    $title: "Width"
                },
                $height: {
                    $type: "application/x-string",
                    $title: "Height"
                }
            }
        },
        $locationTitle: field.$field.$location.$title,
        $locationUrl: field.$field.$location.$url,
        $article: {
            $title: "HTML Vignette",
            $layout: {
                $items: [{
                    $bind: "$locationTitle",
                    $isEditMode: true,
                    $labelWidth: "auto",
                    $skin: "s-author-field"
                }, {
                    $bind: "$locationUrl",
                    $isEditMode: true,
                    $labelWidth: "auto",
                    $skin: "s-author-field"
                }, {
                    $layout: {
                        $layoutType: "side",
                        $items: [{
                            $bind: "$width",
                            $isEditMode: true,
                            $labelWidth: "auto",
                            $skin: "s-author-field"
                        }, {
                            $bind: "$height",
                            $isEditMode: true,
                            $labelWidth: "auto",
                            $skin: "s-author-field"
                        }]
                    }
                }]
            }
        }
    };
};
exports.dispose = function(field){
    document.site.removeResizeListener(field);
};
