// syracuse-ui/lib/mobile/field/vignette/menusVignetteBuilder.js
"use strict";
exports.load = function(field){
    if (field.$field.$location) {
        document.controller.loadRepresentation(null, field.$field.$location.$url, function($itemPage){
            $itemPage.$$container = field.$$body;
            $itemPage.$category = "vignette";
            $itemPage.$skin = $itemPage.$skin || "s-vignette-widget-menus";
            field.vignette = document.itemFactory.loadPage($itemPage);
        });
    }
    else {
        field.vignette = document.itemFactory.loadPage({
            $$container: field.$$body,
            $representation: {
                $prototype: {
                    $links: field.$field.$links,
                    $actions: field.$field.$actions
                },
                $article: {
                    $category: "vignette",
                    $layout: {
                        $items: [{
                            $theme: document.site.$item.$themes.$vignetteMenus,
                            $format: field.$item.$menusFormat || "icon",
                            $category: "links"
                        }]
                    }
                }
            },
            $urlParts: {
                $facet: "$menu"
            },
            $category: "vignette"
        });
    }
};
exports.getAuthoringWidget = function(field){
    return null;
};
exports.dispose = function(){
};
