// syracuse-ui/lib/mobile/field/vignette/vignetteField.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var MobileBox = require("syracuse-ui/lib/mobile/article/box").MobileBox;

var _formatBuilder = {
    "$html": require('./htmlVignetteBuilder'),
    "$page": {
        load: function(field){
            document.controller.loadRepresentation(null, field.$field.$location.$url, function($itemPage){
                $itemPage.$$container = field.$$body;
                $itemPage.$category = "vignette";
                field.vignette = document.itemFactory.loadPage($itemPage);
            });
        },
        getAuthoringWidget: function(field){
            return null;
        },
        dispose: function(){
        }
    },
    "$item": {
        load: function(field){
            field.vignette = document.itemFactory.load(field.$$body, field.$field.$item, field);
        },
        getAuthoringWidget: function(field){
            return null;
        },
        dispose: function(){
        }
    },
    "$menu": require('./menusVignetteBuilder')
};

function MobileVignetteField(){
}

exports.MobileVignetteField = helpers.defineClass(MobileVignetteField, MobileBox, {
    loadBox: function(){
        this.$authoringLevel = "field";
        this.$skin = "s-vignette-field";
        MobileBox.prototype.loadBox.call(this);
    },
    onModalResize: function(){
        // Nothing       
    },
    drawBox: function(){
        var dashboard = this.getPage();
        this.$theme = document.site.$item.$themes.$field.$vignetteField;
        if (this.$item.$isBoxCollapsable === undefined) {
            this.$item.$isBoxCollapsable = false;
        }
        this.$$item = $("<div/>").appendTo(this.$$container);
        if (!this.$item.$isTitleHidden) {
            this.appendHeader(null, false, true);
        }
        this.applyDesignMetaData(this.$item, false);
        this.$$body = $("<div class='" + this.$skin + "-body'/>").appendTo(this.$$item);
        this.openBox(this.$item.$opened !== false)
    },
    _renderLayoutContent: function(){
        var self = this;
        if (self.vignette) {
            self.$$body.empty();
            document.controller.disposeObject(self.vignette);
        }
        if (!this.$field.$format && this.$field.$location) {
            this.$field.$format = this.$field.$location.$type == "html" ? "$html" : "$page";
        }
        if (this.formatBuilder = _formatBuilder[this.$field.$format]) {
            this.formatBuilder.load(this);
        }
        return;
        if (self.$field.$location) {
            document.controller.loadRepresentation(null, self.$field.$location.$url, function($itemPage){
                $itemPage.$$container = self.$$body;
                $itemPage.$category = "vignette";
                self.vignette = document.itemFactory.loadPage($itemPage);
            });
        }
        else {
            switch (self.$field.$format) {
                case "$menu":
                    self.vignette = document.itemFactory.loadPage({
                        $$container: self.$$body,
                        $representation: {
                            $prototype: {
                                $links: self.$field.$links,
                                $actions: self.$field.$actions
                            },
                            $article: {
                                $category: "vignette",
                                $layout: {
                                    $items: [{
                                        $theme: document.site.$item.$themes.$vignetteMenus,
                                        $format: self.$item.$menusFormat || "icon",
                                        $category: "links"
                                    }]
                                }
                            }
                        },
                        $urlParts: {
                            $facet: "$menu"
                        },
                        $category: "vignette"
                    });
                    break;
            }
        }
    },
    dispose: function(){
        this.formatBuilder.dispose(this);
        delete this.formatBuilder;
        MobileBox.prototype.dispose.call(this);
    }
});
