// @sage/syracuse-react/src/test/alert/alertPanelUnits.ts
"use strict";
import { IAlertPanel } from '../../alert/alert';

interface IAlertPanelUnit extends IAlertPanel {
    title: string
}

export const alertPanelUnits: IAlertPanelUnit[] = [{
    title: "diagnose Error",
    diagnoses: [{
        message: "Lorem ipsum dolor sit amet",
        severity: "error"
    }]
}, {
    title: "diagnose Warning",
    diagnoses: [{
        message: "ousd fsdfqsd sdfqsdfqsdfqs fqsdqsdfg",
        severity: "warning"
    }]
},
{
    title: "diagnose info autoHide",
    autoHide: true,
    diagnoses: [{
        message: "oulalaldfdf fsd fqsdfsd fsdfqsd sdfqsdfqsdfqs fqsdqsdfg",
        severity: "info"
    }]
}, {
    title: "diagnose Error + links",
    diagnoses: [{
        message: "An error occured. You may find help by clicking one of following links ",
        severity: "error"
    }],
    links: {
        $link1: {
            $title: "link1",
            $url: "?representation=s-uitest-diagnoses.$test"
        },
        $link2: {
            $title: "link2",
            $url: "?representation=s-uitest-diagnoses.$test"
        },
        $link3: {
            $title: "link3",
            $url: "?representation=s-uitest-diagnoses.$test"
        }
    }
}, {
    title: "diagnose success + links",
    diagnoses: [{
        message: "Resource has been saved",
        severity: "success"
    }],
    links: {
        "$link1": {
            "$title": "Ok",
            "$url": "?representation=s-uitest-diagnoses.$test"
        },
        "$link2": {
            "$title": "Home",
            "$url": "?representation=home.$landing"
        }
    }
}, {
    title: "diagnose Error + stacktrace",
    diagnoses: [{
        message: "this is an error message with a stacktrace",
        severity: "error",
        stackTrace: "class=[java.lang.RuntimeException]\nmess(1)=[java.lang.RuntimeException: java.lang.RuntimeException: java.lang.RuntimeException: java.lang.NullPointerException | RuntimeException | com.sage.x3.bundle.internal.fusion.syndication.sap.CSDataPayloadFusionWins(toJson:57)]\nmess(2)=[java.lang.RuntimeException: java.lang.RuntimeException: java.lang.NullPointerException | RuntimeException | com.sage.x3.bundle.internal.fusion.syndication.sap.CSDataPayloadFusionWin(toJson:151)]\nmess(3)=[java.lang.RuntimeException: java.lang.NullPointerException | RuntimeException | com.sage.x3.bundle.shared.adapi.CAdapiDataSrc(toJson:719)]\nmess(4)=[java.lang.NullPointerException | RuntimeException | com.sage.x3.bundle.shared.adapi.CAdapiDataSrcLine(toJson:233)]\nmess(5)=[NullPointerException | com.sage.x3.bundle.shared.adapi.CAdapiVariable(getValueForJson:199)]\nstack=[com.sage.x3.bundle.internal.fusion.syndication.sap.CSDataPayloadFusionWins(toJson:57)\ncom.sage.x3.bundle.internal.fusion.syndication.sap.CSDataPayloadFusionSAP(toJson:167)\ncom.sage.x3.bundle.internal.fusion.syndication.CSDataPayloadFusionResponse(toJson:78)\ncom.sage.x3.bundle.internal.fusion.syndication.CSDataPayloadFusionResponse(toJson:1)\ncom.sage.x3.bundle.internal.fusion.syndication.CSDataEntryFusion(toJSON:60)\ncom.sage.x3.shared.sdata.syndication.CSyndicationBase(writeToStream:449)\ncom.sage.x3.shared.sdata.syndication.CSyndicationBase(dumpInString:86)\ncom.sage.x3.bundle.internal.fusion.syndication.CSDataEntryFusion(dumpInString:28)\ncom.sage.x3.bundle.internal.fusion.x3.handlerRequest.RequestSvc(execHandler:114)\ncom.sage.x3.bundle.internal.fusion.x3.handlerRequest.AbsRequestSvc(start:44)\ncom.sage.x3.bundle.internal.fusion.x3.FusionClient(execRequest:426)\ncom.sage.x3.bundle.internal.fusion.sdata.adapter.FusionClientAdapter(_execSession:250)\ncom.sage.x3.bundle.internal.fusion.sdata.adapter.FusionClientAdapter(execSession:365)\ncom.sage.x3.bundle.sdata.handler.CContractHandler(putSession:741)\ncom.sage.x3.bundle.sdata.handler.CContractHandler(doPutTreatement:287)\ncom.sage.x3.shared.sdata.util.CSDataAbstractHandler(execute:87)\ncom.sage.x3.bundle.internal.sdata.CSDataContractServiceImpl(doExec:128)\ncom.sage.x3.bundle.internal.sdata.CSDataContractServiceImpl(execRequest:152)\ncom.sage.x3.bundle.internal.sdata.server.CSDataSrvServiceImpl(execRequest:188)\ncom.sage.x3.bundle.internal.sdata.server.CSDataSrvServiceImpl(execRequest:269)\ncom.sage.x3.bundle.internal.sdata.server.request.CHttpHandlerSDataRequest(doRequest:27)\ncom.sage.x3.bundle.http.core.utils.CHandlerRequestDefault(doPut:30)\ncom.sage.x3.bundle.http.core.utils.CHandlerRequest(handle:52)\norg.apache.http.protocol.HttpService(doService:369)\norg.apache.http.protocol.HttpService(handleRequest:290)\ncom.sage.x3.bundle.internal.http.core.CWorkerThread(run:40)]\n"
    }]
}, {
    title: "diagnose multi",
    diagnoses: [{
        message: "Resource has been saved",
        severity: "success"
    }, {
        message: "this is an error message with a stacktrace",
        severity: "error",
        stackTrace: "class=[java.lang.RuntimeException]\nmess(1)=[java.lang.RuntimeException: java.lang.RuntimeException: java.lang.RuntimeException: java.lang.NullPointerException | RuntimeException | com.sage.x3.bundle.internal.fusion.syndication.sap.CSDataPayloadFusionWins(toJson:57)]\nmess(2)=[java.lang.RuntimeException: java.lang.RuntimeException: java.lang.NullPointerException | RuntimeException | com.sage.x3.bundle.internal.fusion.syndication.sap.CSDataPayloadFusionWin(toJson:151)]\nmess(3)=[java.lang.RuntimeException: java.lang.NullPointerException | RuntimeException | com.sage.x3.bundle.shared.adapi.CAdapiDataSrc(toJson:719)]\nmess(4)=[java.lang.NullPointerException | RuntimeException | com.sage.x3.bundle.shared.adapi.CAdapiDataSrcLine(toJson:233)]\nmess(5)=[NullPointerException | com.sage.x3.bundle.shared.adapi.CAdapiVariable(getValueForJson:199)]\nstack=[com.sage.x3.bundle.internal.fusion.syndication.sap.CSDataPayloadFusionWins(toJson:57)\ncom.sage.x3.bundle.internal.fusion.syndication.sap.CSDataPayloadFusionSAP(toJson:167)\ncom.sage.x3.bundle.internal.fusion.syndication.CSDataPayloadFusionResponse(toJson:78)\ncom.sage.x3.bundle.internal.fusion.syndication.CSDataPayloadFusionResponse(toJson:1)\ncom.sage.x3.bundle.internal.fusion.syndication.CSDataEntryFusion(toJSON:60)\ncom.sage.x3.shared.sdata.syndication.CSyndicationBase(writeToStream:449)\ncom.sage.x3.shared.sdata.syndication.CSyndicationBase(dumpInString:86)\ncom.sage.x3.bundle.internal.fusion.syndication.CSDataEntryFusion(dumpInString:28)\ncom.sage.x3.bundle.internal.fusion.x3.handlerRequest.RequestSvc(execHandler:114)\ncom.sage.x3.bundle.internal.fusion.x3.handlerRequest.AbsRequestSvc(start:44)\ncom.sage.x3.bundle.internal.fusion.x3.FusionClient(execRequest:426)\ncom.sage.x3.bundle.internal.fusion.sdata.adapter.FusionClientAdapter(_execSession:250)\ncom.sage.x3.bundle.internal.fusion.sdata.adapter.FusionClientAdapter(execSession:365)\ncom.sage.x3.bundle.sdata.handler.CContractHandler(putSession:741)\ncom.sage.x3.bundle.sdata.handler.CContractHandler(doPutTreatement:287)\ncom.sage.x3.shared.sdata.util.CSDataAbstractHandler(execute:87)\ncom.sage.x3.bundle.internal.sdata.CSDataContractServiceImpl(doExec:128)\ncom.sage.x3.bundle.internal.sdata.CSDataContractServiceImpl(execRequest:152)\ncom.sage.x3.bundle.internal.sdata.server.CSDataSrvServiceImpl(execRequest:188)\ncom.sage.x3.bundle.internal.sdata.server.CSDataSrvServiceImpl(execRequest:269)\ncom.sage.x3.bundle.internal.sdata.server.request.CHttpHandlerSDataRequest(doRequest:27)\ncom.sage.x3.bundle.http.core.utils.CHandlerRequestDefault(doPut:30)\ncom.sage.x3.bundle.http.core.utils.CHandlerRequest(handle:52)\norg.apache.http.protocol.HttpService(doService:369)\norg.apache.http.protocol.HttpService(handleRequest:290)\ncom.sage.x3.bundle.internal.http.core.CWorkerThread(run:40)]\n"
    },
    {
        message: "oulalaldfdf fsd fqsdfsd fsdfqsd sdfqsdfqsdfqs fqsdqsdfg",
        severity: "info"
    },
    {
        message: "autoHide Lorem ipsum dolor sit amet",
        severity: "error"
    }, {

        message: "oulalaldfdf fsd fqsdfsd fsdfqsd sdfqsdfqsdfqs fqsdqsdfg",
        severity: "warning"
    }
    ]
}];
