// @sage/syracuse-react/src/page/misc/dev/devMenusPage.tsx
"use strict";
import * as React from 'react';
import { WebPage } from '../../page';
import Section from '../../../section/section';
import { IPageProps } from '../../page';
import { Link } from '../../../link/link';
import { route } from '../../../route/route';
import { menus } from './menus';

interface IGroup {
    title: string,
    menus: IItem[]
}
interface IItem {
    url: string,
    title: string,
    method: string
}

class Group extends React.Component<{
    title: string
    menus: IItem[],
    onClick: (value: any) => void
}, {}>{
    render() {
        let cols: any[][] = [[], [], [], []];
        let colIndex = 0, recordIndex = 0;
        for (let record of this.props.menus) {
            cols[colIndex++].push(<Link
                key={recordIndex++}
                url={record.url}
                className="s_uitest_mn"
                title={record.title}
                value={record}
                onClick={this.props.onClick} />);
            if (colIndex == cols.length) {
                colIndex = 0;
            }
        }
        return <Section title={this.props.title}>
            <div className="row">
                <div className="col-md-3">{cols[0]}</div>
                <div className="col-md-3">{cols[1]}</div>
                <div className="col-md-3">{cols[2]}</div>
                <div className="col-md-3">{cols[3]}</div>
            </div>
        </Section>;
    }
}

export class DevMenusPage extends React.Component<IPageProps, {}> {
    onClick = (item: IItem) => {
        route.push(item.url);
    }
    render() {
        let groups: any[] = menus.map((group: IGroup, index: number) => {
            return <Group key={index} title={group.title} menus={group.menus} onClick={this.onClick} />
        });
        return <WebPage title="Misc links">
            {groups}
        </WebPage>;
    }
}
