// syracuse-ui/lib/localization/localizer.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var TranslationPage = require('syracuse-ui/lib/localization/translationPage').TranslationPage;


function _ensureLocalizeItemId($item) {
	if (!$item.$title || $item.$title.indexOf("{@") < 0) {
		$item.$title = "{@" + helpers.uuid.generate() + "}";
	}
}

function _extractTitleKey($title) {
	return syra_expression.extractCode($title);
}

function _updatePageLocalization(page, $key, values) {
	var language = exports.getLanguage();
	var $localization = page.$item.$localization = page.$item.$localization || {};
	for (var ii = 0, jj = values.length; ii < jj; ii++) {
		var value = values[ii];
		if (value.$value) {
			($localization[value.$locale] = $localization[value.$locale] || {})[$key] = value.$value;
			if (value.$locale == language) {
				(page.$prototype.$localization = page.$prototype.$localization || {})[$key] = value.$value;
			}
		} else {
			if ($localization[value.$locale]) {
				delete $localization[value.$locale][$key];
				if (!Object.keys($localization[value.$locale]).length) {
					delete $localization[value.$locale];
				}
			}
		}
	}
}


function _setPageLocalizationEntry(page, $key, $value) {
	if (page && page.$item) {
		var $localization = page.$item.$localization = page.$item.$localization || {};
		var language = exports.getLanguage();
		if (language) {
			($localization[language] = $localization[language] || {})[$key] = $value;
			(page.$prototype.$localization = page.$prototype.$localization || {})[$key] = $value;
		}
	}
};

exports.getLanguage = function() {
	return syra_site.userProfile.currentLangCode;
};


exports.getDefaultTabTitle = function(boxParent, $id) {
	var defaultTitle = syra_local.defaultTabTitle;
	var tabs = boxParent.layoutContent && boxParent.layoutContent.items;
	var count = 1;
	if (tabs) {
		for (var ii = 0, jj = tabs.length; ii < jj; ii++) {
			var title = tabs[ii].getTitle() || "";
			if (title.indexOf(defaultTitle) >= 0) {
				count++;
			}
		}
	}
	var $key = "@" + ($id || helpers.uuid.generate());
	_setPageLocalizationEntry(boxParent.page, $key, defaultTitle + " " + count);
	return "{" + $key + "}";
};

exports.applyPageLocalization = function(page) {
	var $pageLocal = page.$item.$localization;
	if ($pageLocal) {
		var langs = Object.keys($pageLocal);
		if (langs.length) {
			var language = exports.getLanguage();

			var $localization;
			if (language) {
				$localization = $pageLocal[language] || $pageLocal[language.toLowerCase()];
			}
			if (!$localization) {
				$localization = $pageLocal["en-us"] || $pageLocal[langs[0]] || $pageLocal[langs[0].toLowerCase()];
			}
			if ($localization) {
				page.$prototype.$localization = page.$prototype.$localization || {};
				syra_delta.applyObjectDelta(page, page.$prototype.$localization, $localization);
			}
			page.$prototype.$allLocalization = $pageLocal;
		}
	}
};


function _onValidateFieldValue(page, field) {
	syra_workingCopy.validateAllInputs(page);
	var $menu = page.$menus.$save;
	if ($menu.$url !== 'test') { // 'test' is $url value for $save link in test page
		var reqOpt = {
			method: $menu.$method,
			$location: {
				$url: $menu.$url
			},
			data: {
				$values: page.$values
			},
			$contentType: "application/json"
		};
		syra_controller.callServer(null, reqOpt, function(data, response, $url) {
			// assuming client receives data for the underneath page
			if (!(field.articleParent.onLocalizeField && !field.articleParent.onLocalizeField(field, data, response, $url))) {
				field.page.applyChange(data);
			}
			page.dialogWrapper && page.dialogWrapper.close(undefined, true);

		}, function(error) {
			var $diagnoses;
			if (error.data.indexOf("$diagnoses") != -1) {
				$diagnoses = JSON.parse(error.data).$diagnoses;
			} else {
				$diagnoses = [{
					$severity: "error",
					$message: error.data
				}];
			}
			syra_diagnose.showDiagnoses({
				$diagnoses: $diagnoses
			}, page);
		});
		return false;
	}
	return true;
}

function _localizeFieldValue(field, $menu) {
	var value = field.getDataValue();
	syra_controller.callServer(null, {
		method: "GET",
		$location: {
			$url: syra_expression.parse(field.articleParent, $menu.$url)
		}
	}, function(data, response, $url) {
		if (!data.$values) {
			syra_diagnose.showDiagnoses({
				$diagnoses: [{
					$message: syra_local.lPageNoValues
				}]
			}, field);
			return false;
		}
		var page = new TranslationPage();
		page.isReadOnly = !(data.$links && data.$links.$save);
		page.loadBox(data);
		var options = {
			page: page
		};
		if (!page.isReadOnly) {
			options.onValidate = function(page) {
				return _onValidateFieldValue(page, field);
			};
		}
		syra_dlg.openPage(field.page, options);
	}, function(error) {
		var $diagnoses;
		if (error.data.indexOf("$diagnoses") != -1) {
			$diagnoses = JSON.parse(error.data).$diagnoses;
		} else {
			$diagnoses = [{
				$severity: "error",
				$message: error.data
			}];
		}
		syra_diagnose.showDiagnoses({
			$diagnoses: $diagnoses
		});
	});
}

function _localizeSectionTitle(field, $menu) {
	syra_site.userProfile.getRegionalPreferences(function(preferences) {
		var $pageLocalization = field.page.designedArticle.$item.$localization || {};
		var $values = [];
		var section = field.localizedSection;
		var $titleKey = _extractTitleKey(section.$item.$title);
		var lcl, lclLower;
		for (var ii = 0, jj = preferences.length; ii < jj; ii++) {
			/*Following SAM 114170, we ignore translation on locale "xx-YY" if translation on locale "xx-yy" exists too.
			 Corollary, if "xx-YY" and "xx-yy" are both sent by the server, we forcibly delete translation entry "xx-YY".
			 If there is only translation on locale "xx-YY", we keep it as is. If there is no translation for one supported  locale (on entry "xx-YY" or "xx-yy"), we set locale code to "xx-yy" for further translation...*/
			lcl = preferences[ii].code;
			lclLower = lcl.toLowerCase();
			var value = {
				$title: preferences[ii].description,
				$locale: lclLower
			};
			if ($pageLocalization[lclLower]) {
				if ($pageLocalization[lcl]) {
					delete field.page.designedArticle.$item.$localization[lcl];
				}
				lcl = lclLower;
			} else if ($pageLocalization[lcl]) {
				value.$locale = lcl;
			}
			if ($pageLocalization[lcl]) {
				value.$value = $pageLocalization[lcl][$titleKey];
			}
			$values.push(value);
		}
		var page = new TranslationPage();
		page.loadBox({
			$values: $values
		});
		syra_dlg.openPage(field.page, {
			page: page,
			onValidate: function(page) {
				syra_workingCopy.validateAllInputs(page);
				_updatePageLocalization(section.page, $titleKey, page.$values);
				section.setTitle(section.$item.$title);
				field.setDataValue(section.getTitle());
				field.page.saveDesign(); //page is designer
				return true;
			}
		});
	});
}

exports.setSectionTitleEditor = function(field, section) {
	field.localizedSection = section;
	_ensureLocalizeItemId(section.$item);
	if (section.isTitleUnlocalized) {
		field.setDataValue("");
	} else {
		field.setDataValue(section.getTitle());
	}
};

exports.onSectionTitleEditorChange = function(field, value) {
	var section = field.localizedSection;
	_setPageLocalizationEntry(section.page, _extractTitleKey(section.$item.$title), value);
	section.setTitle(section.$item.$title);
};
exports.onMenuClick = function(field, $menu) {
	if (field.localizedSection) {
		_localizeSectionTitle(field, $menu);
	} else {
		_localizeFieldValue(field, $menu);
	}
};
