// syracuse-ui/lib/tests/diagnosesTestFk.js
"use strict";
var helpers = require("syracuse-core/lib/helpers")
var RawPage = require("syracuse-ui/lib/article/rawPage").RawPage;
var DiagnosesPanel = require('syracuse-ui/lib/diagnoses/diagnosesPanel').DiagnosesPanel;

function DiagnosesTest(){
}

exports.DiagnosesTest = helpers.defineClass(DiagnosesTest, RawPage, {
    loadBox: function(){
        var self = this;
        self.diagnoseTypes = {
            error: {
                $title: "1 Error",
                $diagnose: {
                    message: "oulalaldfdf fsd fqsdfsd fsdfqsd sdfqsdfqsdfqs fqsdqsdfg",
                    severity: "error"
                }
            },
            fatal: {
                $title: "1 Fatal",
                $diagnose: {
                    message: "oulalaldfdf fsd fqsdfsd fsdfqsd sdfqsdfqsdfqs fqsdqsdfg",
                    severity: "fatal"
                }
            },
            warning: {
                $title: "1 Warning",
                $diagnose: {
                    message: "oulalaldfdf fsd fqsdfsd fsdfqsd sdfqsdfqsdfqs fqsdqsdfg",
                    severity: "warning"
                }
            },
            info: {
                $title: "1 info",
                $diagnose: {
                    message: "oulalaldfdf fsd fqsdfsd fsdfqsd sdfqsdfqsdfqs fqsdqsdfg",
                    severity: "info"
                },
                $links: {
                    $link1: {
                        $title: "link1 bla bla bla"
                    },
                    $link2: {
                        $title: "link2 bla bla bla"
                    },
                    $link3: {
                        $title: "link3 bla bla bla"
                    },
                    $link4: {
                        $title: "link4 bla bla bla"
                    }
                }
            }
        };
        self.$prototype = {
            $links: {}
        };
        Object.keys(self.diagnoseTypes).forEach(function($type){
            self.$prototype.$links[$type] = {
                $title: "Open popup"
            }
        });
        self.$prototype.$links["multitest"] = {
            $title: "Open popup"
        }
        self.$prototype.$links["$cleandiagnoses"] = {
            $title: "Clean Diagnoses"
        }
        self.$item = {
            $skin: "s-vignette",
            $layout: {
                $items: self._makeSection(["error", "fatal", "info", "warning", "multitest"])
            }
        };
        self.$autoFetch = false;
        self.$pageCategory = "vignette";
        RawPage.prototype.loadBox.call(self);
    },
    _makeSection: function($types){
        return $types.map(function($type){
            return {
                $category: "section",
                $id: "diagnoseSection-" + $type,
                $title: "severity = '" + $type + "'",
                $layout: {
                    $items: [{
                        $category: "link",
                        $bind: $type
                    }, {
                        $category: "link",
                        $bind: "$cleandiagnoses"
                    }]
                }
            }
        });
    },
    
    onMenuClick: function(menu){
        if (menu.$bind == "$cleandiagnoses") {
            document.site.showDiagnoses({}, document.site.mainPage);
        }
        else {
            document.site.showDiagnoses(this.buildDiagnose(menu.$bind), document.site.mainPage);
        }
    },
    
    buildDiagnose: function($type){
        var self = this;
        var msg = {
            $diagnoses: []
        };
        if ($type == "multitest") {
            for (var ii = 0; ii < 3; ii++) {
                Object.keys(self.diagnoseTypes).forEach(function($type){
                    msg.$diagnoses.push(self.diagnoseTypes[$type].$diagnose);
                });
            }
        }
        else {
            var diagnoseType = self.diagnoseTypes[$type];
            if (diagnoseType.$links) {
                msg.$links = diagnoseType.$links;
            }
            msg.$diagnoses.push(diagnoseType.$diagnose);
        }
        return msg;
    }
});
