// syracuse-ui/lib/tests/editorFieldTest.js
"use strict";
var helpers = require("syracuse-core/lib/helpers")
var FieldsTest = require("./fieldsTest").FieldsTest;

function EditorFieldTest(){
}

exports.EditorFieldTest = helpers.defineClass(EditorFieldTest, FieldsTest, {
    loadBox: function(){
        FieldsTest.prototype.loadBox.call(this, {
            "$properties": {
                "$field$TestProperty": {
                    $title: "$editMode",
                    $type: "application/x-choice",
                    $value: {
                        $type: "application/x-integer",
                        $enum: [{
                            $value: "array",
                            $title: "array (default)"
                        }, {
                            $value: "record",
                            $title: "record"
                        }, {
                            $value: "cell",
                            $title: "cell"
                        }]
                    }
                },
                "textrtf": {
                    $title: "text-rtf",
                    $type: "text/rtf",
                    
                    $maxLength: 30,
                    $capabilities: "sort,filter,alphaTab",
                    $links: {
                        $tunnel: {},
                        $lookup: {}
                    }
                }
            }
        }, {
            "$field$TestProperty": "record",
            "textrtf": '<ul><li><span style="font-style: italic;"><span style="font-weight: bold;">CSDFSFS</span></span></li><li><span style=" font-style: italic;"><span style=" font-weight: bold;">ZE<span style="color: rgb(255, 0, 0);">FAZR GF</span>ZR</span></span></li><li><span style="  font-style: italic;"><span style="  font-weight: bold;">FRF ZE <br></span></span></li></ul>'
        });
    },
    appendSettingsBar: function(){
        var $items = FieldsTest.prototype.appendSettingsBar.call(this);
        $items.unshift({
            $category: "block",
            $title: "Html Editor Specific",
            $layout: {
                $items: [{
                    $bind: "$field$TestProperty",
                    $isEditMode: true
                }]
            }
        });
        return $items;
    },
    notifyChange: function($bind, value){
        switch ($bind) {
            case "$field$TestProperty":
                alert(value);
                break;
            default:
                FieldsTest.prototype.notifyChange.call(this, $bind, value);
                break;
        }
    },
    _makeBlock: function($isEditMode){
        return [this._makeEditReadBlock("textrtf", [{
            $bind: "textrtf"
        }])];
    }
});
