// syracuse-ui/lib/tests/jobsTest.js
"use strict";
var helpers = require("syracuse-core/lib/helpers")
var TestPage = require("syracuse-ui/lib/tests/testPage").TestPage;
var jobsUtils = require("syracuse-ui/lib/jobs/jobsUtils");

function JobsTest(){
}

exports.JobsTest = helpers.defineClass(JobsTest, TestPage, {
    _initiliazeTestPage: function(){
        var self = this;
        self.jobsTypes = {
            simpleJob: {
                $title: "simple job request test",
                $links: {
                    $link1: {
                        $title: "simple job request test"
                    }
                }
            },
            printJob: {
                $title: "print job request test",
                $links: {
                    $link1: {
                        $title: "print job request test"
                    }
                }
            }
        };
        
        Object.keys(self.jobsTypes).forEach(function($type){
            self.$prototype.$links[$type] = {
                $title: self.jobsTypes[$type].$links.$link1.$title
            }
        });
        
        var cases = ["simpleJob", "printJob"];
        
        
        self.$item.$layout = {
            $items: cases.map(function($type){
                return {
                    $category: "section",
                    $id: "jobSection-" + $type,
                    $title: self.jobsTypes[$type].$title,
                    $layout: {
                        $items: [{
                            $category: "link",
                            $bind: $type
                        }]
                    }
                };
            })
        };
    },
    
    onMenuClick: function(menu){
        var $url;
        switch (menu.$bind) {
            case "simpleJob":
                $url = "http://localhost:8126/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&trackngId=1";
                break;
            case "printJob":
                $url = "http://localhost:8126/print/ecvmdev2003x64:1890/$getState('uuid')";
                break;
        }
        jobsUtils.startService($url);
        return false;
    }
    
});
