// syracuse-ui/lib/tests/jobs.js
"use strict";
var date = require('syracuse-core/lib/types/date');
var time = require('syracuse-core/lib/types/time');

exports.load = function(page) {
	page.leftCol.appendChild(syra_menus.addTextButton("Job = Print request", "s-uitest-mn", "onJobClick"));
	page.setTitle("Jobs");
	page.onJobClick = function(event, target) {
		var startTime = time.now().toString();
		var startDate = date.today().toString();
		syra_jobs.addJob({
			location: "/sdata/$trackers('63023233-9f27-41bb-9749-2331a99596ea')",
			uuid: "record0",
			$state: "Job started",
			phaseDetail: "print",
			progress: 25,
			elapsedSeconds: 10,
			remainingSeconds: 50,
			startTime: startTime,
			startDate: startDate,
			$progressValue: "off",
			title: "Pays",
			kind: "print",
			$links: {
				$download: {
					$title: "Download"
				}
			},
			$diagnoses: [0, 1, 2, 3, 4].map(function(ii) {
				return {
					$message: "number " + ii,
					$severity: "error"
				};
			})
		});

		syra_jobs.addJob({
			uuid: "record1",
			$state: "Job started",
			phaseDetail: "printing",
			progress: 50,
			elapsedSeconds: 10,
			remainingSeconds: 50,
			startTime: startTime,
			startDate: startDate,
			$progressValue: "off",
			title: "Compagnies",
			kind: "print",
			$diagnoses: [{
				$message: "This is a test diagnose : warning",
				$severity: "warning",
				$stackTrace: "TypeError: Object 1115 has no method 'replace' at C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:18:39 at Array.forEach (native) at change (C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:16:20) at C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:20:41 at Array.forEach (native) at change (C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:16:20) at C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:20:41 at Array.forEach (native) at change (C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:16:20) at C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:20:41 at Array.forEach (native) at change (C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:16:20) at C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:20:41 at Array.forEach (native) at change (C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:16:20) at C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:20:41 at Array.forEach (native) at change (C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:16:20) at C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:20:41 at Array.forEach (native) "
			}],
			$links: {
				$cancel: {
					$title: "Cancel"
				}
			}
		});

		syra_jobs.addJob({
			uuid: "record2",
			$state: "Job started",
			phaseDetail: "updating index",
			progress: 50,
			elapsedSeconds: 10,
			remainingSeconds: 50,
			startTime: startTime,
			startDate: startDate,
			$progressValue: "on",
			title: "Index update",
			kind: "operation",
			$links: {
				$abort: {
					$title: "Abort"
				}
			}
		});
	};
};
