// syracuse-ui/lib/tests/jobsTest.js
"use strict";
var helpers = require("syracuse-core/lib/helpers")
var TestPage = require("syracuse-ui/lib/tests/testPage").TestPage;
var jobsUtils = require("syracuse-ui/lib/jobs/jobsUtils");

function JobsTest(){
}

exports.JobsTest = helpers.defineClass(JobsTest, TestPage, {
    _initiliazeTestPage: function(){
        var self = this;
        self.jobsTypes = {
            job: {
                $title: "job request test",
                $links: {
                    $link1: {
                        $title: "job request test"
                    }
                }
            }
        };
        Object.keys(self.jobsTypes).forEach(function($type){
            self.$prototype.$links[$type] = {
                $title: "job request test"
            }
        });
        
        var cases = ["job"];
        
        
        self.$item.$layout = {
            $items: cases.map(function($type){
                return {
                    $category: "section",
                    $id: "jobSection-" + $type,
                    $title: "JOB REQUEST TEST",//"severity = '" + self.jobsTypes[$type].$title + "'",
                    $layout: {
                        $items: [{
                            $category: "link",
                            $bind: $type
                        }]
                    }
                };
            })
        };
    },
    
    onMenuClick: function(menu){
        jobsUtils.startService();
        return false;
    }
    
});
