// syracuse-ui/lib/tests/layoutTest.js
"use strict";
var helpers = require("syracuse-core/lib/helpers")
var RawPage = require("syracuse-ui/lib/article/rawPage").RawPage;

function LayoutTest(){
}

exports.LayoutTest = helpers.defineClass(LayoutTest, RawPage, {
    loadBox: function(){
        var self = this;
        self.$prototype = {
            "$properties": {
                $field$isHidden: {
                    $type: "application/x-boolean",
                    $title: "$isHidden",
                    $isMetaData: true
                },
                $field$isTitleHidden: {
                    $type: "application/x-boolean",
                    $title: "$isTitleHidden",
                    $isMetaData: true
                },
                $field$isBoxCollapsable: {
                    $type: "application/x-boolean",
                    $title: "$isBoxCollapsable",
                    $isMetaData: true
                },
                $field$isMaximizable: {
                    $type: "application/x-boolean",
                    $title: "$isMaximizable ",
                    $isMetaData: true
                }
            }
        };
        
        self.$item = {
            $skin: "s-vignette",
            $layout: {
                $layoutType: "columns",
                $layoutSubType: "70,30",
                $items: [{
                    $layoutType: "stack",
                    $items: ["Page", "Dashboard", "Vignette"].map(function($title){
                        var $layoutOptions = document.site.$item.$layoutOptions.page[$title.toLowerCase()];
                        return {
                            $layoutType: "columns",
                            $layoutSubType: "33,33,33",
                            $items: [["stack", null, null], ["stack", "columns", "tabs"], ["tabs", "tabs", null]].map(function($layouts){
                                return {
                                    $layoutType: $layouts[0],
                                    $items: [self._makeLevelPart($title, $layoutOptions, $layouts[1]), self._makeLevelPart($title, $layoutOptions, $layouts[2])]
                                };
                            })
                        };
                    })
                }, {
                    $layoutType: "stack",
                    $items: [{
                        $category: "section",
                        $title: "Status",
                        $layout: {
                            $items: [{
                                $category: "block",
                                $title: "Metadata",
                                $layout: {
                                    $items: [{
                                        $bind: "$field$isHidden",
                                        $isEditMode: true
                                    }, {
                                        $bind: "$field$isTitleHidden",
                                        $isEditMode: true
                                    }, {
                                        $bind: "$field$isBoxCollapsable",
                                        $isEditMode: true
                                    }, {
                                        $bind: "$field$isMaximizable",
                                        $isEditMode: true
                                    }]
                                }
                            }]
                        }
                    }]
                }]
            }
        };
        self.$autoFetch = false;
        self.$pageCategory = "vignette";
        RawPage.prototype.loadBox.call(self);
    },
    
    notifyChange: function($bind, value){
        var self = this;
        if ($bind.indexOf("$field") == 0) {
            var $field = self.$prototype.$properties[$bind];
            $bind = $bind.replace("$field", "");
            var metaData = {};
            metaData[$bind] = value;
            Object.keys(self.idMap).forEach(function(id){
                var box = self.idMap[id];
                if (box.$item.$test) {
                    box.applyDesignMetaData(metaData, true);
                }
            });
        }
    },
    _makeLevelPart: function($title, $layoutOptions, $layoutType){
        var section = {
            $category: "section",
            $test: true,
            $skin: $layoutOptions.section,
            $title: "Section " + $title,
            $layout: {
                $layoutType: $layoutType || "stack",
                $items: [{
                    $category: "block",
                    $test: true,
                    $skin: $layoutOptions.block,
                    $title: "Block " + $title,
                    $layout: {
                        $items: []
                    }
                }, {
                    $category: "block",
                    $test: true,
                    $skin: $layoutOptions.block,
                    $title: "Block " + $title,
                    $layout: {
                        $items: []
                    }
                }]
            }
        };
        if ($layoutType == "columns") {
            var oldLayout = section.$layout;
            section.$layout = {
                $layoutType: "columns",
                $layoutSubType: "50,50",
                $items: [{
                    $layoutType: "stack",
                    $items: [oldLayout.$items[0]]
                }, {
                    $layoutType: "stack",
                    $items: [oldLayout.$items[1]]
                }]
            };
        }
        return section
    },
    _makeLevelTab: function($title, $items){
        return {
            $layoutType: "tabs",
            $items: [this._makeLevelPart(), this._makeLevelPart()]
        };
    }
});
