// syracuse-ui/lib/tests/menusTest.js
"use strict";
var helpers = require("syracuse-core/lib/helpers")
var DesktopPage = require("syracuse-ui/lib/article/page").DesktopPage;
var resources = require("./resources");

function MenusTest(){
}

exports.MenusTest = helpers.defineClass(MenusTest, DesktopPage, {
    loadBox: function(){
        var self = this;
        self.$prototype.$actions = {};
        ["$links", "$actions", "$linksParameters", "$actionsParameters"].forEach(function($value){
            var $menus = resources.getActionLinks($value);
            ["$links", "$actions"].forEach(function($type){
                if ($menus[$type]) {
                    Object.keys($menus[$type]).forEach(function($bind){
                        self.$prototype[$type][$bind] = $menus[$type][$bind];
                    });
                }
            });
        });
        
        self.$item = {
            $layout: {
                $items: [{
                    $category: "section",
                    $title: "s-page-menus",
                    $layout: {
                        $items: [{
                            $category: "links",
                            $id: "s-page-menus",
                            $skin: "s-page-menus",
                            $isBindDisabled: true,
                            $links: helpers.object.clone(self.$prototype.$links),
                            $actions: helpers.object.clone(self.$prototype.$actions)
                        }]
                    }
                }, {
                    $category: "section",
                    $title: "s-vignette-menus",
                    $layout: {
                        $items: [{
                            $category: "links",
                            $skin: "s-vignette-menus",
                            $isBindDisabled: true,
                            $links: self.$prototype.$links,
                            $actions: self.$prototype.$actions
                        }]
                    }
                }, {
                    $category: "section",
                    $title: "s-vignette-widget-menus",
                    $layout: {
                        $items: [{
                            $category: "links",
                            $skin: "s-vignette-widget-menus",
                            $isBindDisabled: true,
                            $links: self.$prototype.$links,
                            $actions: self.$prototype.$actions
                        }]
                    }
                }, {
                    $category: "section",
                    $title: "s-list-menus (popup)",
                    $layout: {
                        $items: [{
                            $category: "links",
                            $skin: "s-list-menus",
                            $isBindDisabled: true,
                            $links: self.$prototype.$links,
                            $actions: self.$prototype.$actions
                        }]
                    }
                }, {
                    $category: "section",
                    $title: "s-record-menus (popup)",
                    $layout: {
                        $items: [{
                            $category: "links",
                            $skin: "s-record-menus",
                            $isBindDisabled: true,
                            $links: self.$prototype.$links,
                            $actions: self.$prototype.$actions
                        }]
                    }
                }, {
                    $category: "section",
                    $title: "s-field-menus (popup)",
                    $layout: {
                        $items: [{
                            $category: "links",
                            $skin: "s-field-menus",
                            $isBindDisabled: true,
                            $links: self.$prototype.$links,
                            $actions: self.$prototype.$actions
                        }]
                    }
                }, {
                    $category: "section",
                    $title: "s-site-header-menus",
                    $layout: {
                        $items: [{
                            $category: "links",
                            $skin: "s-site-header-menus",
                            $isBindDisabled: true,
                            $links: self.$prototype.$links,
                            $actions: self.$prototype.$actions
                        }]
                    }
                }, {
                    $category: "section",
                    $title: "s-site-top-pn-nav-bar-menus",
                    $layout: {
                        $items: [{
                            $category: "links",
                            $id: "s-site-top-pn-nav-bar-menus",
                            $skin: "s-site-top-pn-nav-bar-menus",
                            $isBindDisabled: true,
                            $links: self.$prototype.$links,
                            $actions: self.$prototype.$actions
                        }]
                    }
                }, {
                    $category: "section",
                    $title: "s-site-top-pn-about-menus",
                    $layout: {
                        $items: [{
                            $category: "links",
                            $skin: "s-site-top-pn-about-menus",
                            $isBindDisabled: true,
                            $links: self.$prototype.$links,
                            $actions: self.$prototype.$actions
                        }]
                    }
                }]
            }
        };
        self.$autoFetch = false;
        DesktopPage.prototype.loadBox.call(self);
        self.idMap["s-page-menus"].$$item[0].style.position = "relative";
        self.idMap["s-site-top-pn-nav-bar-menus"].boxParent.$$body[0].style.backgroundColor = "rgba(0, 0, 0, 0.7)";
    }
    
});
