// syracuse-ui/lib/tests/menusTest.js
"use strict";
var helpers = require("syracuse-core/lib/helpers")
var RawPage = require("syracuse-ui/lib/article/rawPage").RawPage;
var resources = require("./resources");

function MenusTest(){
}

exports.MenusTest = helpers.defineClass(MenusTest, RawPage, {
    loadBox: function(){
        var self = this;
        self.$prototype = {
            $links: {},
            $actions: {}
        };
        ["$links", "$actions", "$linksParameters", "$actionsParameters"].forEach(function($value){
            var $menus = resources.getActionLinks($value);
            ["$links", "$actions"].forEach(function($type){
                if ($menus[$type]) {
                    Object.keys($menus[$type]).forEach(function($bind){
                        self.$prototype[$type][$bind] = $menus[$type][$bind];
                    });
                }
            });
        });
        
        self.$item = {
            $skin: "s-vignette",
            $layout: {
                $items: [{
                    $category: "section",
                    $title: "s-page-menus",
                    $layout: {
                        $items: [{
                            $category: "links",
                            $skin: "s-page-menus",
                            $isBindDisabled: true,
                            $links: helpers.object.clone(self.$prototype.$links),
                            $actions: helpers.object.clone(self.$prototype.$actions)
                        }]
                    }
                },{
                    $category: "section",
                    $title: "s-vignette-menus",
                    $layout: {
                        $items: [{
                            $category: "links",
                            $skin: "s-vignette-menus",
                            $isBindDisabled: true,
                            $links: self.$prototype.$links,
                            $actions: self.$prototype.$actions
                        }]
                    }
                },{
                    $category: "section",
                    $title: "s-list-menus",
                    $layout: {
                        $items: [{
                            $category: "links",
                            $skin: "s-list-menus",
                            $isBindDisabled: true,
                            $links: self.$prototype.$links,
                            $actions: self.$prototype.$actions
                        }]
                    }
                },{
                    $category: "section",
                    $title: "s-record-menus",
                    $layout: {
                        $items: [{
                            $category: "links",
                            $skin: "s-record-menus",
                            $isBindDisabled: true,
                            $links: self.$prototype.$links,
                            $actions: self.$prototype.$actions
                        }]
                    }
                },{
                    $category: "section",
                    $title: "s-field-menus",
                    $layout: {
                        $items: [{
                            $category: "links",
                            $skin: "s-field-menus",
                            $isBindDisabled: true,
                            $links: self.$prototype.$links,
                            $actions: self.$prototype.$actions
                        }]
                    }
                }]
				
            }
        };
        self.$autoFetch = false;
        self.$pageCategory = "vignette";
        RawPage.prototype.loadBox.call(self);
    }
    
});
