// syracuse-ui/lib/tests/messageBoxTestFk.js
"use strict";
var helpers = require("syracuse-core/lib/helpers")
var RawPage = require("syracuse-ui/lib/article/rawPage").RawPage;
var locale = require('syracuse-core/lib/locale');

function MessageBoxTest(){
}

exports.MessageBoxTest = helpers.defineClass(MessageBoxTest, RawPage, {
    loadBox: function(){
        var self = this;
        var localize = locale.resources(module)();
        
        var diagnoses = localize.diagnoses;
        var message;
        
        //console.log(self.detail2String(localize.diagnoses));
        
        
        diagnoses.forEach(function(diagnose){
            message += diagnose.stackTrace + "\n\n";
        });
        /*
         self.messageTypes = {
         alert: {
         $title: "This is an alert",
         $message: message
         }
         };*//*
        self.messageTypes = {
            alert: {
                $title: "This is an alert",
                $message: {
                    errMsg: "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed dolor odio, adipiscing dignissim vehicula porttitor, rhoncus a quam. Phasellus tempor. \nLorem ipsum dolor sit amet, consectetur adipiscing elit.",
                    diagnoses: diagnoses
                }
            }
        };*/
		// long
		//Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nunc ut mi neque. Duis non diam eu nisi aliquet dignissim vitae non tortor. Quisque luctus est quis elit hendrerit tempor. In ornare augue turpis, vel viverra magna. Phasellus molestie, libero lobortis iaculis lobortis, ipsum dui ultrices nunc, sed pulvinar lacus tortor ac nunc. Cras rhoncus risus vitae nisl laoreet in eleifend lorem malesuada. Cras tellus ante, lobortis sit amet posuere at, interdum hendrerit ipsum. Praesent nec elit turpis. Nam sit amet augue diam. In a ipsum ut lorem euismod fermentum sit amet in nibh.\nSuspendisse rhoncus odio a urna sollicitudin vel hendrerit enim molestie. Class aptent taciti sociosqu ad litora torquent per conubia nostra, per inceptos himenaeos. Ut sit amet lacus vel quam eleifend condimentum in quis sem. Phasellus nec lacus erat. Sed tristique pharetra turpis nec vehicula. Vivamus sodales lectus urna, quis laoreet urna. Nunc sit amet interdum neque.
		self.messageTypes = {
            alert: {
                $title: "This is an alert",
                $message: "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed dolor odio, adipiscing dignissim vehicula porttitor, rhoncus a quam. Phasellus tempor. \nLorem ipsum dolor sit amet, consectetur adipiscing elit.",
				$stackTrace: diagnoses
            }
        };
        self.$prototype = {
            $links: {}
        };
        Object.keys(self.messageTypes).forEach(function($type){
            self.$prototype.$links[$type] = {
                $title: "Open popup"
            }
        });
        self.$item = {
            $skin: "s-vignette",
            $layout: {
                $items: self._makeSection(["alert"])
            }
        }
        self.$autoFetch = false;
        self.$pageCategory = "vignette";
        RawPage.prototype.loadBox.call(self);
    },
    _makeSection: function($types){
        return $types.map(function($type){
            return {
                $category: "section",
                $id: "messageBoxSection-" + $type,
                $title: "$type = '" + $type + "'",
                $layout: {
                    $items: [{
                        $category: "link",
                        $bind: $type
                    }]
                }
            }
        });
    },
    onMenuClick: function(menu){
        document.site.showMessage(this.messageTypes[menu.$bind]);
    },
    detail2String: function(o){
        if (o == undefined) 
            return "";
        try {
            if (typeof(o) == 'string') {
                if (o.length == 0) 
                    return o;
                o = JSON.parse(o);
            }
            return JSON.stringify(o, null, "   ");
        } 
        catch (e) {
            return o.toString();
        }
    }
    
});
