// syracuse-ui/lib/tests/messageBoxTestFk.js
"use strict";
var helpers = require("syracuse-core/lib/helpers")
var RawPage = require("syracuse-ui/lib/article/rawPage").RawPage;
var locale = require('syracuse-core/lib/locale');

function MessageBoxTest(){
}

var _titles = {
    "ex1": "alert + links",
    "ex2": "warning + origin",
    "ex3": "fatal + origin + details",
    "ex4": "alert + origin + details + diagnoses"
}

exports.MessageBoxTest = helpers.defineClass(MessageBoxTest, RawPage, {
    loadBox: function(){
        var self = this;
        var localize = locale.resources(module)();
        
        var diagnoses = localize.diagnoses;
        
        // long
        //Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nunc ut mi neque. Duis non diam eu nisi aliquet dignissim vitae non tortor. Quisque luctus est quis elit hendrerit tempor. In ornare augue turpis, vel viverra magna. Phasellus molestie, libero lobortis iaculis lobortis, ipsum dui ultrices nunc, sed pulvinar lacus tortor ac nunc. Cras rhoncus risus vitae nisl laoreet in eleifend lorem malesuada. Cras tellus ante, lobortis sit amet posuere at, interdum hendrerit ipsum. Praesent nec elit turpis. Nam sit amet augue diam. In a ipsum ut lorem euismod fermentum sit amet in nibh.\nSuspendisse rhoncus odio a urna sollicitudin vel hendrerit enim molestie. Class aptent taciti sociosqu ad litora torquent per conubia nostra, per inceptos himenaeos. Ut sit amet lacus vel quam eleifend condimentum in quis sem. Phasellus nec lacus erat. Sed tristique pharetra turpis nec vehicula. Vivamus sodales lectus urna, quis laoreet urna. Nunc sit amet interdum neque.
        self.messageTypes = {
            ex1: {
                $type: "alert",
                $title: "This is an alert",
                $message: "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nunc ut mi neque. Duis non diam eu nisi aliquet",
                $links: {
                    $link1: {
                        $title: "link 3",
                        $url: "http://beta.sage.fr"
                    },
                    $link2: {
                        $title: "link 2 bla ",
                        $url: "http://www.sage.fr"
                    },
                    $link3: {
                        $title: "link 1",
                        $url: "http://www.sage.fr"
                    }
                }
            },
            ex2: {
                $type: "warning",
                $title: "This is a warning",
                $message: "An error occured during the process",
                $origin: "Bridge"
            },
            ex3: {
                $type: "fatal",
                $title: "This is a fatal",
                $message: "A fatal error occured",
                $origin: "x3",
                $details: "Endpoint unavailable.\nUpdate in progress"
            },
            ex4: {
                $type: "alert",
                $title: "This is an alert",
                $message: "Exceptions raised",
                $origin: "Bridge",
                $details: "Try catch has worked",
                $diagnoses: diagnoses
            }
        };
        self.$prototype = {
            $links: {}
        };
        Object.keys(self.messageTypes).forEach(function($type){
            self.$prototype.$links[$type] = {
                $title: "Open popup"
            }
        });
        self.$item = {
            $skin: "s-vignette",
            $layout: {
                $items: [{
                    $layoutType: "stack",
                    $items: self._makeSection(["ex1"])
                }, {
                    $layoutType: "stack",
                    $items: self._makeSection(["ex2"])
                }, {
                    $layoutType: "stack",
                    $items: self._makeSection(["ex3"])
                }, {
                    $layoutType: "stack",
                    $items: self._makeSection(["ex4"])
                }]
            }
        }
        RawPage.prototype.loadBox.call(self);
    },
    _makeSection: function($types){
        return $types.map(function($type){
            return {
                $category: "section",
                $id: "messageBoxSection-" + $type,
                $title: _titles[$type],
                $layout: {
                    $items: [{
                        $category: "link",
                        $bind: $type
                    }]
                }
            }
        });
    },
    onMenuClick: function(menu){
        document.site.showMessage(this.messageTypes[menu.$bind]);
        // lose focus
        menu.$$item.blur();
    }
});
