// syracuse-ui/lib/tests/messageBoxTestFk.js
"use strict";
var helpers = require("syracuse-core/lib/helpers")
var RawPage = require("syracuse-ui/lib/article/rawPage").RawPage;
var locale = require('syracuse-core/lib/locale');

function MessageBoxTest(){
}

exports.MessageBoxTest = helpers.defineClass(MessageBoxTest, RawPage, {
    loadBox: function(){
        var self = this;
        var localize = locale.resources(module)();
        
        var diagnoses = localize.diagnoses;
        var message;
        
        console.log(self.detail2String(localize));
        
        
        diagnoses.forEach(function(diagnose){
            message += diagnose.stackTrace + "\n\n";
        });
        
        self.messageTypes = {
            alert: {
                $title: "This is an alert",
                $message: message
            }
        };
        self.$prototype = {
            $links: {}
        };
        Object.keys(self.messageTypes).forEach(function($type){
            self.$prototype.$links[$type] = {
                $title: "Open popup"
            }
        });
        self.$item = {
            $skin: "s-vignette",
            $layout: {
                $items: self._makeSection(["alert"])
            }
        }
        self.$autoFetch = false;
        self.$pageCategory = "vignette";
        RawPage.prototype.loadBox.call(self);
    },
    _makeSection: function($types){
        return $types.map(function($type){
            return {
                $category: "section",
                $id: "messageBoxSection-" + $type,
                $title: "$type = '" + $type + "'",
                $layout: {
                    $items: [{
                        $category: "link",
                        $bind: $type
                    }]
                }
            }
        });
    },
    onMenuClick: function(menu){
        document.site.showMessage(this.messageTypes[menu.$bind]);
    },
    detail2String: function(o){
        if (o == undefined) 
            return "";
        try {
            if (typeof(o) == 'string') {
                if (o.length == 0) 
                    return o;
                o = JSON.parse(o);
            }
            return JSON.stringify(o, null, "   ");
        } 
        catch (e) {
            return o.toString();
        }
    }
    
});
