// syracuse-ui/lib/tests/messageBoxTest.js
"use strict";
var helpers = require("syracuse-core/lib/helpers")
var TestPage = require("syracuse-ui/lib/tests/testPage").TestPage;
var inlineSettings = require("./messageBoxInlineSettings");
var popupSettings = require("./messageBoxPopupSettings");

function MessageBoxTestInline(){
}

exports.MessageBoxTestInline = helpers.defineClass(MessageBoxTestInline, TestPage, {
    loadBox: function(){
        this.messageTypes = inlineSettings.messageTypes;
        this.$item.$layout = inlineSettings.$item.$layout;
        for (var ii = 0, jj = this.messageTypes.length; ii < jj; ii++) {
            this.$prototype.$links[this.messageTypes[ii].$type] = {
                $title: "Open popup"
            };
        }
        TestPage.prototype.loadBox.call(this);
        for (var ii = 0, jj = this.messageTypes.length; ii < jj; ii++) {
            var msg = this.messageTypes[ii];
            var section = this.idMap["messageBoxSection-" + msg.$type];
            if (msg.$type == "info") {
                section.domTitle.textContent += " ($autoClose = 1000)";
            }
            if (msg.$type == "fatal") {
                section.domTitle.textContent += " ($autoClose = true 4s)";
            }
            
            msg.layoutSlot = section.$$body[0];
            document.site.showMessage(msg);
            delete msg.layoutSlot;
        }
    },
    onMenuClick: function(menu){
        document.site.openDialog({
            $itemPage: {
                "$urlParts": {
                    "uri": "/trans/x3/erp/superv/$sessions",
                    "params": {
                        "f": "GESAMK%2F2%2F%2FM%2F",
                        "representation": "EXEFNC.$fusion"
                    },
                    "target": "trans",
                    "endpoint": "x3.erp.superv",
                    "isFusion": true,
                    "endpointParts": {
                        "application": "x3",
                        "contract": "erp",
                        "dataset": "superv"
                    },
                    "$facet": "$fusion_$modal2"
                },
                "$representation": {
                    "$authorUrl": "http://localhost:8124/sdata/syracuse/collaboration/syracuse/pageAuths/$template/$workingCopies?representation=pageAuth.$edit&pageContext=x3.erp.EXEFNC.$fusion&device=desktop",
                    "$url": "x3.erp.superv.EXEFNC.$fusion,trans,",
                    "$prototype": {
                        "$type": "application/json",
                        "$title": "{@W-EXEFNC-23655}",
                        "$x3Facet": "dialog",
                        "$properties": {
                            "AA1": {
                                "$type": "application/x-icon",
                                "$url": "",
                                "$X3Name": "EXEFNC_F1_1",
                                "$links": {},
                                "$title": "",
                                "$X3Fmt": "KIB:3X",
                                "$maxLength": 3,
                                "$groupName": "A"
                            },
                            "AA2": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_FONC1_1",
                                "$title": "",
                                "$isReadOnly": true,
                                "$X3Fmt": "KPTD:12B",
                                "$maxLength": 12,
                                "$groupName": "A"
                            },
                            "AA3": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_LIBF1_1",
                                "$X3Fmt": "KPW:30X>5X",
                                "$maxLength": 35,
                                "$displayLength": 30,
                                "$groupName": "A"
                            },
                            "AA4": {
                                "$type": "application/x-icon",
                                "$url": "",
                                "$X3Name": "EXEFNC_F1_2",
                                "$links": {},
                                "$X3Fmt": "KIB:3X",
                                "$maxLength": 3,
                                "$groupName": "A"
                            },
                            "AA5": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_FONC1_2",
                                "$isReadOnly": true,
                                "$X3Fmt": "KPTD:12B",
                                "$maxLength": 12,
                                "$groupName": "A"
                            },
                            "AA6": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_LIBF1_2",
                                "$X3Fmt": "KPW:30X>5X",
                                "$maxLength": 35,
                                "$displayLength": 30,
                                "$groupName": "A"
                            },
                            "AA7": {
                                "$type": "application/x-icon",
                                "$url": "",
                                "$X3Name": "EXEFNC_F1_3",
                                "$links": {},
                                "$X3Fmt": "KIB:3X",
                                "$maxLength": 3,
                                "$groupName": "A"
                            },
                            "AA8": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_FONC1_3",
                                "$isReadOnly": true,
                                "$X3Fmt": "KPTD:12B",
                                "$maxLength": 12,
                                "$groupName": "A"
                            },
                            "AA9": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_LIBF1_3",
                                "$X3Fmt": "KPW:30X>5X",
                                "$maxLength": 35,
                                "$displayLength": 30,
                                "$groupName": "A"
                            },
                            "AA10": {
                                "$type": "application/x-icon",
                                "$url": "",
                                "$X3Name": "EXEFNC_F1_4",
                                "$links": {},
                                "$X3Fmt": "KIB:3X",
                                "$maxLength": 3,
                                "$groupName": "A"
                            },
                            "AA11": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_FONC1_4",
                                "$isReadOnly": true,
                                "$X3Fmt": "KPTD:12B",
                                "$maxLength": 12,
                                "$groupName": "A"
                            },
                            "AA12": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_LIBF1_4",
                                "$X3Fmt": "KPW:30X>5X",
                                "$maxLength": 35,
                                "$displayLength": 30,
                                "$groupName": "A"
                            },
                            "AA13": {
                                "$type": "application/x-icon",
                                "$url": "",
                                "$X3Name": "EXEFNC_F1_5",
                                "$links": {},
                                "$X3Fmt": "KIB:3X",
                                "$maxLength": 3,
                                "$groupName": "A"
                            },
                            "AA14": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_FONC1_5",
                                "$isReadOnly": true,
                                "$X3Fmt": "KPTD:12B",
                                "$maxLength": 12,
                                "$groupName": "A"
                            },
                            "AA15": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_LIBF1_5",
                                "$X3Fmt": "KPW:30X>5X",
                                "$maxLength": 35,
                                "$displayLength": 30,
                                "$groupName": "A"
                            },
                            "AA16": {
                                "$type": "application/x-icon",
                                "$url": "",
                                "$X3Name": "EXEFNC_F1_6",
                                "$links": {},
                                "$X3Fmt": "KIB:3X",
                                "$maxLength": 3,
                                "$groupName": "A"
                            },
                            "AA17": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_FONC1_6",
                                "$isReadOnly": true,
                                "$X3Fmt": "KPTD:12B",
                                "$maxLength": 12,
                                "$groupName": "A"
                            },
                            "AA18": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_LIBF1_6",
                                "$X3Fmt": "KPW:30X>5X",
                                "$maxLength": 35,
                                "$displayLength": 30,
                                "$groupName": "A"
                            },
                            "AA19": {
                                "$type": "application/x-icon",
                                "$url": "",
                                "$X3Name": "EXEFNC_F1_7",
                                "$links": {},
                                "$X3Fmt": "KIB:3X",
                                "$maxLength": 3,
                                "$groupName": "A"
                            },
                            "AA20": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_FONC1_7",
                                "$isReadOnly": true,
                                "$X3Fmt": "KPTD:12B",
                                "$maxLength": 12,
                                "$groupName": "A"
                            },
                            "AA21": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_LIBF1_7",
                                "$X3Fmt": "KPW:30X>5X",
                                "$maxLength": 35,
                                "$displayLength": 30,
                                "$groupName": "A"
                            },
                            "AA22": {
                                "$type": "application/x-icon",
                                "$url": "",
                                "$X3Name": "EXEFNC_F1_8",
                                "$links": {},
                                "$X3Fmt": "KIB:3X",
                                "$maxLength": 3,
                                "$groupName": "A"
                            },
                            "AA23": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_FONC1_8",
                                "$isReadOnly": true,
                                "$X3Fmt": "KPTD:12B",
                                "$maxLength": 12,
                                "$groupName": "A"
                            },
                            "AA24": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_LIBF1_8",
                                "$X3Fmt": "KPW:30X>5X",
                                "$maxLength": 35,
                                "$displayLength": 30,
                                "$groupName": "A"
                            },
                            "AB1": {
                                "$type": "application/x-icon",
                                "$url": "",
                                "$X3Name": "EXEFNC_F2_1",
                                "$links": {},
                                "$title": "",
                                "$X3Fmt": "KIB:3X",
                                "$maxLength": 3,
                                "$groupName": "B"
                            },
                            "AB2": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_FONC2_1",
                                "$title": "",
                                "$isReadOnly": true,
                                "$X3Fmt": "KPTD:12B",
                                "$maxLength": 12,
                                "$groupName": "B"
                            },
                            "AB3": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_LIBF2_1",
                                "$X3Fmt": "KPW:30X>5X",
                                "$maxLength": 35,
                                "$displayLength": 30,
                                "$groupName": "B"
                            },
                            "AB4": {
                                "$type": "application/x-icon",
                                "$url": "",
                                "$X3Name": "EXEFNC_F2_2",
                                "$links": {},
                                "$X3Fmt": "KIB:3X",
                                "$maxLength": 3,
                                "$groupName": "B"
                            },
                            "AB5": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_FONC2_2",
                                "$isReadOnly": true,
                                "$X3Fmt": "KPTD:12B",
                                "$maxLength": 12,
                                "$groupName": "B"
                            },
                            "AB6": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_LIBF2_2",
                                "$X3Fmt": "KPW:30X>5X",
                                "$maxLength": 35,
                                "$displayLength": 30,
                                "$groupName": "B"
                            },
                            "AB7": {
                                "$type": "application/x-icon",
                                "$url": "",
                                "$X3Name": "EXEFNC_F2_3",
                                "$links": {},
                                "$X3Fmt": "KIB:3X",
                                "$maxLength": 3,
                                "$groupName": "B"
                            },
                            "AB8": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_FONC2_3",
                                "$isReadOnly": true,
                                "$X3Fmt": "KPTD:12B",
                                "$maxLength": 12,
                                "$groupName": "B"
                            },
                            "AB9": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_LIBF2_3",
                                "$X3Fmt": "KPW:30X>5X",
                                "$maxLength": 35,
                                "$displayLength": 30,
                                "$groupName": "B"
                            },
                            "AB10": {
                                "$type": "application/x-icon",
                                "$url": "",
                                "$X3Name": "EXEFNC_F2_4",
                                "$links": {},
                                "$X3Fmt": "KIB:3X",
                                "$maxLength": 3,
                                "$groupName": "B"
                            },
                            "AB11": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_FONC2_4",
                                "$isReadOnly": true,
                                "$X3Fmt": "KPTD:12B",
                                "$maxLength": 12,
                                "$groupName": "B"
                            },
                            "AB12": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_LIBF2_4",
                                "$X3Fmt": "KPW:30X>5X",
                                "$maxLength": 35,
                                "$displayLength": 30,
                                "$groupName": "B"
                            },
                            "AB13": {
                                "$type": "application/x-icon",
                                "$url": "",
                                "$X3Name": "EXEFNC_F2_5",
                                "$links": {},
                                "$X3Fmt": "KIB:3X",
                                "$maxLength": 3,
                                "$groupName": "B"
                            },
                            "AB14": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_FONC2_5",
                                "$isReadOnly": true,
                                "$X3Fmt": "KPTD:12B",
                                "$maxLength": 12,
                                "$groupName": "B"
                            },
                            "AB15": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_LIBF2_5",
                                "$X3Fmt": "KPW:30X>5X",
                                "$maxLength": 35,
                                "$displayLength": 30,
                                "$groupName": "B"
                            },
                            "AB16": {
                                "$type": "application/x-icon",
                                "$url": "",
                                "$X3Name": "EXEFNC_F2_6",
                                "$links": {},
                                "$X3Fmt": "KIB:3X",
                                "$maxLength": 3,
                                "$groupName": "B"
                            },
                            "AB17": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_FONC2_6",
                                "$isReadOnly": true,
                                "$X3Fmt": "KPTD:12B",
                                "$maxLength": 12,
                                "$groupName": "B"
                            },
                            "AB18": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_LIBF2_6",
                                "$X3Fmt": "KPW:30X>5X",
                                "$maxLength": 35,
                                "$displayLength": 30,
                                "$groupName": "B"
                            },
                            "AB19": {
                                "$type": "application/x-icon",
                                "$url": "",
                                "$X3Name": "EXEFNC_F2_7",
                                "$links": {},
                                "$X3Fmt": "KIB:3X",
                                "$maxLength": 3,
                                "$groupName": "B"
                            },
                            "AB20": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_FONC2_7",
                                "$isReadOnly": true,
                                "$X3Fmt": "KPTD:12B",
                                "$maxLength": 12,
                                "$groupName": "B"
                            },
                            "AB21": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_LIBF2_7",
                                "$X3Fmt": "KPW:30X>5X",
                                "$maxLength": 35,
                                "$displayLength": 30,
                                "$groupName": "B"
                            },
                            "AB22": {
                                "$type": "application/x-icon",
                                "$url": "",
                                "$X3Name": "EXEFNC_F2_8",
                                "$links": {},
                                "$X3Fmt": "KIB:3X",
                                "$maxLength": 3,
                                "$groupName": "B"
                            },
                            "AB23": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_FONC2_8",
                                "$isReadOnly": true,
                                "$X3Fmt": "KPTD:12B",
                                "$maxLength": 12,
                                "$groupName": "B"
                            },
                            "AB24": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_LIBF2_8",
                                "$X3Fmt": "KPW:30X>5X",
                                "$maxLength": 35,
                                "$displayLength": 30,
                                "$groupName": "B"
                            },
                            "AB25": {
                                "$type": "application/x-icon",
                                "$url": "",
                                "$X3Name": "EXEFNC_F2_9",
                                "$links": {},
                                "$X3Fmt": "KIB:3X",
                                "$maxLength": 3,
                                "$groupName": "B"
                            },
                            "AB26": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_FONC2_9",
                                "$isReadOnly": true,
                                "$X3Fmt": "KPTD:12B",
                                "$maxLength": 12,
                                "$groupName": "B"
                            },
                            "AB27": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_LIBF2_9",
                                "$X3Fmt": "KPW:30X>5X",
                                "$maxLength": 35,
                                "$displayLength": 30,
                                "$groupName": "B"
                            },
                            "AB28": {
                                "$type": "application/x-icon",
                                "$url": "",
                                "$X3Name": "EXEFNC_F2_10",
                                "$links": {},
                                "$X3Fmt": "KIB:3X",
                                "$maxLength": 3,
                                "$groupName": "B"
                            },
                            "AB29": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_FONC2_10",
                                "$isReadOnly": true,
                                "$X3Fmt": "KPTD:12B",
                                "$maxLength": 12,
                                "$groupName": "B"
                            },
                            "AB30": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_LIBF2_10",
                                "$X3Fmt": "KPW:30X>5X",
                                "$maxLength": 35,
                                "$displayLength": 30,
                                "$groupName": "B"
                            },
                            "AB31": {
                                "$type": "application/x-icon",
                                "$url": "",
                                "$X3Name": "EXEFNC_F2_11",
                                "$links": {},
                                "$X3Fmt": "KIB:3X",
                                "$maxLength": 3,
                                "$groupName": "B"
                            },
                            "AB32": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_FONC2_11",
                                "$isReadOnly": true,
                                "$X3Fmt": "KPTD:12B",
                                "$maxLength": 12,
                                "$groupName": "B"
                            },
                            "AB33": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_LIBF2_11",
                                "$X3Fmt": "KPW:30X>5X",
                                "$maxLength": 35,
                                "$displayLength": 30,
                                "$groupName": "B"
                            },
                            "AB34": {
                                "$type": "application/x-icon",
                                "$url": "",
                                "$X3Name": "EXEFNC_F2_12",
                                "$links": {},
                                "$X3Fmt": "KIB:3X",
                                "$maxLength": 3,
                                "$groupName": "B"
                            },
                            "AB35": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_FONC2_12",
                                "$isReadOnly": true,
                                "$X3Fmt": "KPTD:12B",
                                "$maxLength": 12,
                                "$groupName": "B"
                            },
                            "AB36": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_LIBF2_12",
                                "$X3Fmt": "KPW:30X>5X",
                                "$maxLength": 35,
                                "$displayLength": 30,
                                "$groupName": "B"
                            },
                            "AB37": {
                                "$type": "application/x-icon",
                                "$url": "",
                                "$X3Name": "EXEFNC_F2_13",
                                "$links": {},
                                "$X3Fmt": "KIB:3X",
                                "$maxLength": 3,
                                "$groupName": "B"
                            },
                            "AB38": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_FONC2_13",
                                "$isReadOnly": true,
                                "$X3Fmt": "KPTD:12B",
                                "$maxLength": 12,
                                "$groupName": "B"
                            },
                            "AB39": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_LIBF2_13",
                                "$X3Fmt": "KPW:30X>5X",
                                "$maxLength": 35,
                                "$displayLength": 30,
                                "$groupName": "B"
                            },
                            "AB40": {
                                "$type": "application/x-icon",
                                "$url": "",
                                "$X3Name": "EXEFNC_F2_14",
                                "$links": {},
                                "$X3Fmt": "KIB:3X",
                                "$maxLength": 3,
                                "$groupName": "B"
                            },
                            "AB41": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_FONC2_14",
                                "$isReadOnly": true,
                                "$X3Fmt": "KPTD:12B",
                                "$maxLength": 12,
                                "$groupName": "B"
                            },
                            "AB42": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_LIBF2_14",
                                "$X3Fmt": "KPW:30X>5X",
                                "$maxLength": 35,
                                "$displayLength": 30,
                                "$groupName": "B"
                            },
                            "AB43": {
                                "$type": "application/x-icon",
                                "$url": "",
                                "$X3Name": "EXEFNC_F2_15",
                                "$links": {},
                                "$X3Fmt": "KIB:3X",
                                "$maxLength": 3,
                                "$groupName": "B"
                            },
                            "AB44": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_FONC2_15",
                                "$isReadOnly": true,
                                "$X3Fmt": "KPTD:12B",
                                "$maxLength": 12,
                                "$groupName": "B"
                            },
                            "AB45": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_LIBF2_15",
                                "$X3Fmt": "KPW:30X>5X",
                                "$maxLength": 35,
                                "$displayLength": 30,
                                "$groupName": "B"
                            },
                            "AB46": {
                                "$type": "application/x-icon",
                                "$url": "",
                                "$X3Name": "EXEFNC_F2_16",
                                "$links": {},
                                "$X3Fmt": "KIB:3X",
                                "$maxLength": 3,
                                "$groupName": "B"
                            },
                            "AB47": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_FONC2_16",
                                "$isReadOnly": true,
                                "$X3Fmt": "KPTD:12B",
                                "$maxLength": 12,
                                "$groupName": "B"
                            },
                            "AB48": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_LIBF2_16",
                                "$X3Fmt": "KPW:30X>5X",
                                "$maxLength": 35,
                                "$displayLength": 30,
                                "$groupName": "B"
                            },
                            "AC1": {
                                "$type": "application/x-icon",
                                "$url": "",
                                "$X3Name": "EXEFNC_FFO",
                                "$links": {},
                                "$title": "",
                                "$X3Fmt": "KIB:3X",
                                "$maxLength": 3,
                                "$groupName": "C"
                            },
                            "AC2": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_FONCTION",
                                "$valueStyle": "AFVSEL",
                                "$titleStyle": "AFTSEL",
                                "$links": {
                                    "$lookup": {},
                                    "$lazyload": {}
                                },
                                "$title": "",
                                "$X3Fmt": "KTD:12B",
                                "$maxLength": 12,
                                "$groupName": "C"
                            },
                            "AC3": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_LIBFON",
                                "$X3Fmt": "KPW:30X>5X",
                                "$maxLength": 35,
                                "$displayLength": 30,
                                "$groupName": "C"
                            },
                            "AD1": {
                                "$type": "application/x-icon",
                                "$url": "",
                                "$X3Name": "EXEFNC_FFR",
                                "$links": {},
                                "$title": "",
                                "$X3Fmt": "KIB:3X",
                                "$maxLength": 3,
                                "$groupName": "D"
                            },
                            "AD2": {
                                "$type": "application/x-string",
                                "$X3Name": "EXEFNC_RECHFCT",
                                "$title": "",
                                "$X3Fmt": "K:50X>30X",
                                "$maxLength": 80,
                                "$displayLength": 50,
                                "$groupName": "D"
                            }
                        },
                        "$article": {
                            "$layout": {
                                "$layoutType": "stack",
                                "$items": [{
                                    "$title": "{@W-EXEFNC--10001601}",
                                    "$category": "section",
                                    "$XID": "A",
                                    "$layout": {
                                        "$layoutType": "stack",
                                        "$items": [{
                                            "$layoutType": "row",
                                            "$items": [{
                                                "$title": "",
                                                "$category": "block",
                                                "$layout": {
                                                    "$items": [{
                                                        "$layoutType": "row",
                                                        "$items": [{
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": true,
                                                            "$bind": "AA1"
                                                        }, {
                                                            "$isReadOnly": true,
                                                            "$bind": "AA2"
                                                        }, {
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": false,
                                                            "$isTopLabelAlignment": true,
                                                            "$bind": "AA3"
                                                        }]
                                                    }, {
                                                        "$layoutType": "row",
                                                        "$items": [{
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": true,
                                                            "$bind": "AA4"
                                                        }, {
                                                            "$isReadOnly": true,
                                                            "$bind": "AA5"
                                                        }, {
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": false,
                                                            "$isTopLabelAlignment": true,
                                                            "$bind": "AA6"
                                                        }]
                                                    }, {
                                                        "$layoutType": "row",
                                                        "$items": [{
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": true,
                                                            "$bind": "AA7"
                                                        }, {
                                                            "$isReadOnly": true,
                                                            "$bind": "AA8"
                                                        }, {
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": false,
                                                            "$isTopLabelAlignment": true,
                                                            "$bind": "AA9"
                                                        }]
                                                    }, {
                                                        "$layoutType": "row",
                                                        "$items": [{
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": true,
                                                            "$bind": "AA10"
                                                        }, {
                                                            "$isReadOnly": true,
                                                            "$bind": "AA11"
                                                        }, {
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": false,
                                                            "$isTopLabelAlignment": true,
                                                            "$bind": "AA12"
                                                        }]
                                                    }, {
                                                        "$layoutType": "row",
                                                        "$items": [{
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": true,
                                                            "$bind": "AA13"
                                                        }, {
                                                            "$isReadOnly": true,
                                                            "$bind": "AA14"
                                                        }, {
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": false,
                                                            "$isTopLabelAlignment": true,
                                                            "$bind": "AA15"
                                                        }]
                                                    }, {
                                                        "$layoutType": "row",
                                                        "$items": [{
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": true,
                                                            "$bind": "AA16"
                                                        }, {
                                                            "$isReadOnly": true,
                                                            "$bind": "AA17"
                                                        }, {
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": false,
                                                            "$isTopLabelAlignment": true,
                                                            "$bind": "AA18"
                                                        }]
                                                    }, {
                                                        "$layoutType": "row",
                                                        "$items": [{
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": true,
                                                            "$bind": "AA19"
                                                        }, {
                                                            "$isReadOnly": true,
                                                            "$bind": "AA20"
                                                        }, {
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": false,
                                                            "$isTopLabelAlignment": true,
                                                            "$bind": "AA21"
                                                        }]
                                                    }, {
                                                        "$layoutType": "row",
                                                        "$items": [{
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": true,
                                                            "$bind": "AA22"
                                                        }, {
                                                            "$isReadOnly": true,
                                                            "$bind": "AA23"
                                                        }, {
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": false,
                                                            "$isTopLabelAlignment": true,
                                                            "$bind": "AA24"
                                                        }]
                                                    }]
                                                }
                                            }, {
                                                "$title": "",
                                                "$category": "block",
                                                "$layout": {
                                                    "$items": [{
                                                        "$layoutType": "row",
                                                        "$items": [{
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": true,
                                                            "$bind": "AB1"
                                                        }, {
                                                            "$isReadOnly": true,
                                                            "$bind": "AB2"
                                                        }, {
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": false,
                                                            "$isTopLabelAlignment": true,
                                                            "$bind": "AB3"
                                                        }]
                                                    }, {
                                                        "$layoutType": "row",
                                                        "$items": [{
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": true,
                                                            "$bind": "AB4"
                                                        }, {
                                                            "$isReadOnly": true,
                                                            "$bind": "AB5"
                                                        }, {
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": false,
                                                            "$isTopLabelAlignment": true,
                                                            "$bind": "AB6"
                                                        }]
                                                    }, {
                                                        "$layoutType": "row",
                                                        "$items": [{
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": true,
                                                            "$bind": "AB7"
                                                        }, {
                                                            "$isReadOnly": true,
                                                            "$bind": "AB8"
                                                        }, {
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": false,
                                                            "$isTopLabelAlignment": true,
                                                            "$bind": "AB9"
                                                        }]
                                                    }, {
                                                        "$layoutType": "row",
                                                        "$items": [{
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": true,
                                                            "$bind": "AB10"
                                                        }, {
                                                            "$isReadOnly": true,
                                                            "$bind": "AB11"
                                                        }, {
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": false,
                                                            "$isTopLabelAlignment": true,
                                                            "$bind": "AB12"
                                                        }]
                                                    }, {
                                                        "$layoutType": "row",
                                                        "$items": [{
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": true,
                                                            "$bind": "AB13"
                                                        }, {
                                                            "$isReadOnly": true,
                                                            "$bind": "AB14"
                                                        }, {
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": false,
                                                            "$isTopLabelAlignment": true,
                                                            "$bind": "AB15"
                                                        }]
                                                    }, {
                                                        "$layoutType": "row",
                                                        "$items": [{
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": true,
                                                            "$bind": "AB16"
                                                        }, {
                                                            "$isReadOnly": true,
                                                            "$bind": "AB17"
                                                        }, {
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": false,
                                                            "$isTopLabelAlignment": true,
                                                            "$bind": "AB18"
                                                        }]
                                                    }, {
                                                        "$layoutType": "row",
                                                        "$items": [{
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": true,
                                                            "$bind": "AB19"
                                                        }, {
                                                            "$isReadOnly": true,
                                                            "$bind": "AB20"
                                                        }, {
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": false,
                                                            "$isTopLabelAlignment": true,
                                                            "$bind": "AB21"
                                                        }]
                                                    }, {
                                                        "$layoutType": "row",
                                                        "$items": [{
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": true,
                                                            "$bind": "AB22"
                                                        }, {
                                                            "$isReadOnly": true,
                                                            "$bind": "AB23"
                                                        }, {
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": false,
                                                            "$isTopLabelAlignment": true,
                                                            "$bind": "AB24"
                                                        }]
                                                    }, {
                                                        "$layoutType": "row",
                                                        "$items": [{
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": true,
                                                            "$bind": "AB25"
                                                        }, {
                                                            "$isReadOnly": true,
                                                            "$bind": "AB26"
                                                        }, {
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": false,
                                                            "$isTopLabelAlignment": true,
                                                            "$bind": "AB27"
                                                        }]
                                                    }, {
                                                        "$layoutType": "row",
                                                        "$items": [{
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": true,
                                                            "$bind": "AB28"
                                                        }, {
                                                            "$isReadOnly": true,
                                                            "$bind": "AB29"
                                                        }, {
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": false,
                                                            "$isTopLabelAlignment": true,
                                                            "$bind": "AB30"
                                                        }]
                                                    }, {
                                                        "$layoutType": "row",
                                                        "$items": [{
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": true,
                                                            "$bind": "AB31"
                                                        }, {
                                                            "$isReadOnly": true,
                                                            "$bind": "AB32"
                                                        }, {
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": false,
                                                            "$isTopLabelAlignment": true,
                                                            "$bind": "AB33"
                                                        }]
                                                    }, {
                                                        "$layoutType": "row",
                                                        "$items": [{
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": true,
                                                            "$bind": "AB34"
                                                        }, {
                                                            "$isReadOnly": true,
                                                            "$bind": "AB35"
                                                        }, {
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": false,
                                                            "$isTopLabelAlignment": true,
                                                            "$bind": "AB36"
                                                        }]
                                                    }, {
                                                        "$layoutType": "row",
                                                        "$items": [{
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": true,
                                                            "$bind": "AB37"
                                                        }, {
                                                            "$isReadOnly": true,
                                                            "$bind": "AB38"
                                                        }, {
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": false,
                                                            "$isTopLabelAlignment": true,
                                                            "$bind": "AB39"
                                                        }]
                                                    }, {
                                                        "$layoutType": "row",
                                                        "$items": [{
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": true,
                                                            "$bind": "AB40"
                                                        }, {
                                                            "$isReadOnly": true,
                                                            "$bind": "AB41"
                                                        }, {
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": false,
                                                            "$isTopLabelAlignment": true,
                                                            "$bind": "AB42"
                                                        }]
                                                    }, {
                                                        "$layoutType": "row",
                                                        "$items": [{
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": true,
                                                            "$bind": "AB43"
                                                        }, {
                                                            "$isReadOnly": true,
                                                            "$bind": "AB44"
                                                        }, {
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": false,
                                                            "$isTopLabelAlignment": true,
                                                            "$bind": "AB45"
                                                        }]
                                                    }, {
                                                        "$layoutType": "row",
                                                        "$items": [{
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": true,
                                                            "$bind": "AB46"
                                                        }, {
                                                            "$isReadOnly": true,
                                                            "$bind": "AB47"
                                                        }, {
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": false,
                                                            "$isTopLabelAlignment": true,
                                                            "$bind": "AB48"
                                                        }]
                                                    }]
                                                }
                                            }]
                                        }, {
                                            "$layoutType": "row",
                                            "$items": [{
                                                "$title": "",
                                                "$category": "block",
                                                "$layout": {
                                                    "$items": [{
                                                        "$layoutType": "row",
                                                        "$items": [{
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": true,
                                                            "$bind": "AC1"
                                                        }, {
                                                            "$isEditMode": true,
                                                            "$bind": "AC2"
                                                        }, {
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": false,
                                                            "$isTopLabelAlignment": true,
                                                            "$bind": "AC3"
                                                        }]
                                                    }]
                                                }
                                            }]
                                        }, {
                                            "$layoutType": "row",
                                            "$items": [{
                                                "$title": "",
                                                "$category": "block",
                                                "$layout": {
                                                    "$items": [{
                                                        "$layoutType": "row",
                                                        "$items": [{
                                                            "$isEditMode": false,
                                                            "$isTitleHidden": true,
                                                            "$bind": "AD1"
                                                        }, {
                                                            "$isEditMode": true,
                                                            "$bind": "AD2"
                                                        }]
                                                    }]
                                                }
                                            }]
                                        }]
                                    }
                                }]
                            },
                            "$menus": {
                                "$layout": {
                                    "$items": [{
                                        "$style": "main",
                                        "$layout": {
                                            "$items": [{
                                                "$bind": 1115,
                                                "$style": "main"
                                            }, {
                                                "$bind": "2053",
                                                "rnk": 50
                                            }, {
                                                "$bind": 2825,
                                                "rnk": 51
                                            }, {
                                                "$bind": "2845",
                                                "rnk": 52
                                            }]
                                        },
                                        "rnk": 1,
                                        "$id": "997"
                                    }, {
                                        "$bind": 1186,
                                        "rnk": 2
                                    }, {
                                        "$title": "",
                                        "$layout": {
                                            "$items": [{
                                                "$bind": 2823,
                                                "rnk": 1,
                                                "$icon": {
                                                    "$mode": "icon",
                                                    "$path": "x3icons/",
                                                    "$value": "s-vignette-refresh"
                                                }
                                            }, {
                                                "$bind": 2818,
                                                "rnk": 80,
                                                "$icon": {
                                                    "$mode": "icon",
                                                    "$path": "x3icons/",
                                                    "$value": "238"
                                                }
                                            }, {
                                                "$bind": 3356,
                                                "rnk": 81,
                                                "$icon": {
                                                    "$mode": "icon",
                                                    "$path": "x3icons/",
                                                    "$value": "304"
                                                },
                                                "$isHidden": true
                                            }, {
                                                "$bind": 2819,
                                                "rnk": 90,
                                                "$icon": {
                                                    "$mode": "icon",
                                                    "$path": "x3icons/",
                                                    "$value": "260"
                                                }
                                            }, {
                                                "$bind": 3355,
                                                "rnk": 91,
                                                "$icon": {
                                                    "$mode": "icon",
                                                    "$path": "x3icons/",
                                                    "$value": "305"
                                                },
                                                "$isHidden": true
                                            }],
                                            "$layoutType": "row",
                                            "$autoSize": true
                                        },
                                        "rnk": 3,
                                        "$id": "996"
                                    }, {
                                        "$title": "Utilitaires",
                                        "$layout": {
                                            "$items": [{
                                                "$bind": 2817,
                                                "rnk": 20
                                            }, {
                                                "$bind": 2847,
                                                "rnk": 21
                                            }, {
                                                "$bind": 2849,
                                                "rnk": 22
                                            }, {
                                                "$bind": 2822,
                                                "rnk": 50
                                            }, {
                                                "$bind": 2848,
                                                "rnk": 51
                                            }, {
                                                "$bind": 1169,
                                                "rnk": 55
                                            }, {
                                                "$bind": 2844,
                                                "rnk": 56
                                            }, {
                                                "$title": "Imprimer",
                                                "$layout": {
                                                    "$items": [{
                                                        "$bind": 2820,
                                                        "rnk": 10
                                                    }, {
                                                        "$bind": 2821,
                                                        "rnk": 20
                                                    }],
                                                    "$layoutType": "row",
                                                    "$autoSize": true
                                                },
                                                "rnk": 65,
                                                "$opened": false,
                                                "$id": "319",
                                                "$isPopupContent": true,
                                                "$skin": "s-test-menus-icon"
                                            }, {
                                                "$title": "Exporter",
                                                "$layout": {
                                                    "$items": [{
                                                        "$bind": 1174,
                                                        "rnk": 10,
                                                        "$icon": {
                                                            "$mode": "iconText",
                                                            "$path": "file/",
                                                            "$value": "s-link-syracuse-word-mailmerge"
                                                        }
                                                    }, {
                                                        "$bind": 1185,
                                                        "rnk": 20,
                                                        "$icon": {
                                                            "$mode": "iconText",
                                                            "$path": "file/",
                                                            "$value": "s-link-syracuse-excel-worksheet"
                                                        }
                                                    }],
                                                    "$layoutType": "row",
                                                    "$autoSize": true
                                                },
                                                "rnk": 67,
                                                "$opened": false,
                                                "$id": "18000",
                                                "$isPopupContent": true,
                                                "$skin": "s-test-menus-icon"
                                            }]
                                        },
                                        "rnk": 9090,
                                        "$opened": false,
                                        "$id": "322"
                                    }, {
                                        "$title": "Affichage",
                                        "$layout": {
                                            "$items": []
                                        },
                                        "rnk": 9095,
                                        "$opened": false,
                                        "$id": "324"
                                    }, {
                                        "$title": "Navigation",
                                        "$layout": {
                                            "$items": []
                                        },
                                        "rnk": 9097,
                                        "$id": "320"
                                    }, {
                                        "$title": "Aide",
                                        "$layout": {
                                            "$items": [{
                                                "$bind": 2846,
                                                "rnk": 12
                                            }, {
                                                "$bind": 1168,
                                                "rnk": 41
                                            }, {
                                                "$title": "Aide au diagnostic",
                                                "$layout": {
                                                    "$items": [{
                                                        "$bind": 1041,
                                                        "rnk": 11,
                                                        "$icon": {
                                                            "$mode": "iconText",
                                                            "$path": "x3icons/",
                                                            "$value": "42"
                                                        }
                                                    }, {
                                                        "$bind": 1040,
                                                        "rnk": 12,
                                                        "$icon": {
                                                            "$mode": "iconText",
                                                            "$path": "x3icons/",
                                                            "$value": "282"
                                                        }
                                                    }, {
                                                        "$bind": 1172,
                                                        "rnk": 21
                                                    }, {
                                                        "$bind": 1173,
                                                        "rnk": 31
                                                    }, {
                                                        "$bind": 1183,
                                                        "rnk": 41
                                                    }, {
                                                        "$bind": 1184,
                                                        "rnk": 42
                                                    }, {
                                                        "$bind": 1165,
                                                        "rnk": 51
                                                    }, {
                                                        "$bind": 1166,
                                                        "rnk": 52
                                                    }]
                                                },
                                                "rnk": 51,
                                                "$opened": false,
                                                "$id": "318"
                                            }]
                                        },
                                        "rnk": 9100,
                                        "$opened": false,
                                        "$id": "323"
                                    }]
                                }
                            }
                        },
                        "$localization": {
                            "@W-EXEFNC-23655": "Excution d'une fonction",
                            "@W-EXEFNC--100157": "OK",
                            "@W-EXEFNC-25208": "Ajout favoris",
                            "@W-EXEFNC--123065": "Fin",
                            "@W-EXEFNC--10001601": "Page 1"
                        },
                        "$role": "f56e5600-a09d-4d4c-83fe-161c9ff13db2",
                        "$representationUrl": "http://localhost:8124/trans/x3/erp/superv/$sessions?f=GESAMK%2F2%2F%2FM%2F",
                        "$links": {
                            "2053": {
                                "$act": "2053",
                                "$title": "{@W-EXEFNC--100157}"
                            },
                            "2825": {
                                "$act": 2825,
                                "$title": "{@W-EXEFNC-25208}"
                            },
                            "2845": {
                                "$act": "2845",
                                "$title": "{@W-EXEFNC--123065}"
                            },
                            "1186": {
                                "$act": 1186,
                                "$title": "Aller  la fonction..."
                            },
                            "2823": {
                                "$act": 2823,
                                "$title": "Actualiser ..."
                            },
                            "2818": {
                                "$act": 2818,
                                "$title": "Pices jointes"
                            },
                            "3356": {
                                "$act": 3356,
                                "$title": "Pices jointes"
                            },
                            "2819": {
                                "$act": 2819,
                                "$title": "Commentaires"
                            },
                            "3355": {
                                "$act": 3355,
                                "$title": "Commentaires"
                            },
                            "2817": {
                                "$act": 2817,
                                "$title": "Changer la cl..."
                            },
                            "2847": {
                                "$act": 2847,
                                "$title": "Statistiques"
                            },
                            "2849": {
                                "$act": 2849,
                                "$title": "Liaisons"
                            },
                            "2822": {
                                "$act": 2822,
                                "$title": "Proprits..."
                            },
                            "2848": {
                                "$act": 2848,
                                "$title": "Envoyer  ..."
                            },
                            "1169": {
                                "$act": 1169,
                                "$title": "Convertisseur"
                            },
                            "2844": {
                                "$act": 2844,
                                "$title": "Choisir Transaction ..."
                            },
                            "2820": {
                                "$act": 2820,
                                "$title": "Fiche"
                            },
                            "2821": {
                                "$act": 2821,
                                "$title": "Liste"
                            },
                            "1174": {
                                "$act": 1174,
                                "$title": "Donnes en XML"
                            },
                            "1185": {
                                "$act": 1185,
                                "$title": "Donnes vers Excel"
                            },
                            "2846": {
                                "$act": 2846,
                                "$title": "Aide sur fiche"
                            },
                            "1168": {
                                "$act": 1168,
                                "$title": "Proprits"
                            },
                            "1041": {
                                "$act": 1041,
                                "$title": "Calculatrice"
                            },
                            "1040": {
                                "$act": 1040,
                                "$title": "Debugger"
                            },
                            "1172": {
                                "$act": 1172,
                                "$title": "Informations du champ"
                            },
                            "1173": {
                                "$act": 1173,
                                "$title": "Dbg On/Off"
                            },
                            "1183": {
                                "$act": 1183,
                                "$title": "Activation timing"
                            },
                            "1184": {
                                "$act": 1184,
                                "$title": "Lecture timing"
                            },
                            "1165": {
                                "$act": 1165,
                                "$title": "Activation trace Oracle"
                            },
                            "1166": {
                                "$act": 1166,
                                "$title": "Dsactivation trace"
                            },
                            "1115": {
                                "$title": "Nouveau",
                                "$act": 1115
                            }
                        }
                    },
                    "$uuid": "629ed4cd-25ba-4701-ace1-154372640272",
                    "$key": "629ed4cd-25ba-4701-ace1-154372640272"
                },
                "$autoFetch": false,
                "$category": "fusion",
                "$isEditMode": true,
                "boxParent": null
            }
        });
        
        for (var ii = 0, jj = this.messageTypes.length; ii < jj; ii++) {
            var msg = this.messageTypes[ii];
            if (msg.$type == menu.$bind) {
                if (msg.cbk && !msg.callback) {
                    msg.callback = function(response){
                        if (response) {
                            alert(response[msg.$type == "prompt" ? "prompt" : "$id"]);
                        }
                    };
                }
                document.site.showMessage(msg);
                return false;
            }
        }
        return true;
    }
});

function MessageBoxTestPopup(){
}

exports.MessageBoxTestPopup = helpers.defineClass(MessageBoxTestPopup, TestPage, {
    _initiliazeTestPage: function(){
        this.messageTypes = popupSettings.messageTypes;
        this.$item.$layout = popupSettings.$item.$layout;
        this.$prototype = popupSettings.$prototype;
    },
    onMenuClick: function(menu){
        document.site.showMessage(this.messageTypes[parseInt(menu.$bind, 10) - 1]);
        // lose focus
        //menu.$$item.blur();
    }
});
