// syracuse-ui/lib/tests/resources.js
"use strict";
exports.getCube = function() {
	return [{
		"$uuid": "d3e97372-21c1-40ac-b147-4a0eaad4f58e",
		"$key": "d3e97372-21c1-40ac-b147-4a0eaad4f58e",
		"ARTICLE": "Toto",
		"code": "TOOT",
		"matiere_premiere": "10.3",
		"main_oeuvre": "20.5",
		"cout_machine": "56.3"
	}, {
		"$uuid": "a3e97372-21c1-40ac-b147-4a0eaad4f58e",
		"$key": "a3e97372-21c1-40ac-b147-4a0eaad4f58e",
		"ARTICLE": "Cle USB",
		"code": "CLEUSB",
		"matiere_premiere": "12.3",
		"main_oeuvre": "45.2",
		"cout_machine": "125.2"
	}, {
		"$uuid": "b3e97372-21c1-40ad-b147-4a0eaad4f58e",
		"$key": "b3e97372-21c1-40ad-b147-4a0eaad4f58e",
		"ARTICLE": "Ecran Plasma",
		"code": "ECRPLASMA",
		"matiere_premiere": "125.2",
		"main_oeuvre": "50.3",
		"cout_machine": "100.1"
	}, {
		"$uuid": "c3e97372-21c2-40ac-b147-4a0eaad4f58e",
		"$key": "c3e97372-21c2-40ac-b147-4a0eaad4f58e",
		"ARTICLE": "Lecteur DVD",
		"code": "Lecteur DVD",
		"matiere_premiere": "45.2",
		"main_oeuvre": "23.3",
		"cout_machine": "22.3"
	}, {
		"$uuid": "d4e97342-21c1-40ac-b147-4a0eaad4f58e",
		"$key": "d4e97342-21c1-40ac-b147-4a0eaad4f58e",
		"ARTICLE": "Imprimante Laser",
		"code": "IMPLASER",
		"matiere_premiere": "45.2",
		"main_oeuvre": "65.00",
		"cout_machine": "45.25"
	}, {
		"$uuid": "d5e97372-22c1-40ac-b147-4a0eaad4f58e",
		"$key": "d5e97372-22c1-40ac-b147-4a0eaad4f58e",
		"ARTICLE": "CD-ROM",
		"code": "CDROM",
		"matiere_premiere": "2.0",
		"main_oeuvre": "12.32",
		"cout_machine": "10.5"
	}];
};
exports.getRecords = function(count) {
	var records = [];
	for (var index = 0; index < count; index++) {
		records.push({
			"$uuid": "record" + index,
			"$etag": 1,
			"$properties": {},
			"$index": index,
			"xstring": "record " + index,
			"xchoice": index,
			"xdate": "2011-0" + Math.max(Math.min(1, index), 9) + "-07",
			"xboolean": true,
			"xdecimal": 45454.0 * index,
			"xreference": {
				"$uuid": "4e54a41e-5c7d-4926-ad57-6a5f882c24c4",
				"$key": "4e54a41e-5c7d-4926-ad57-6a5f882c24c4",
				"code": "AD",
				"description": "Andorradfd fsdqfsqdfsdq sdqf sdqfsdqf sdqqfsdq fsdqf sdqf sdfsdqf sdf sqdfsd qfsdq fsdqf sdqfsd qfsdq fsqd fs q fqsf",
				"$title": "title"
			}
		});
	}
	return records;
};

exports.getActionLinks = function(value) {
	return {
		$empty: {
			$links: null
		},
		$lazyload: {
			$links: {
				"$lazyload": {}
			}
		},
		$links: {
			$links: {
				"$tunnel": {
					"$title": "$tunnel"
				},
				"$lookup": {
					"$title": "$lookup"
				},
				"google": {
					"$title": "Google",
					"$url": "http://www.google.fr/"
				},
				"uioverview": {
					"$title": "UI Overview",
					"$url": "?representation=s-uitest-fields.$test"
				},
				"uilist": {
					"$title": "UI List Overview",
					"$url": "?representation=uitestlist.$dashboard"
				},
				"transitionTest": {
					"$title": "Json Page test",
					"$url": "?representation=uitestpagetest.$dashboard"
				},
				"$create": {
					"$type": "application/json;vnd.sage=syracuse",
					"$title": "Créer",
					"$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries/$template/$workingCopies?representation=country.$edit&role={$role}",
					"$method": "POST"
				},
				"$details": {
					"$type": "application/json;vnd.sage=syracuse",
					"$title": "Detail",
					"$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
				},
				"$edit": {
					"$type": "application/json;vnd.sage=syracuse",
					"$title": "Edit",
					"$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')/$workingCopies?representation=country.$edit&role={$role}",
					"$method": "POST"
				},
				"$delete": {
					"$title": "Delete",
					"$confirm": "The record  will be deleted. Confirm ?",
					"$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$query&role={$role}",
					"$type": "application/json;vnd.sage=syracuse",
					"$method": "DELETE"
				}
			}
		},
		$actions: {
			$actions: {
				$save: {
					$title: "Save"
				}
			}
		},
		$linksParameters: {
			$links: {
				$linksExpressionParameters: {
					$title: "Expression parameters",
					$description: "param2 from xstring,param2 from json",
					$url: "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&param1={param1}&param2={param2}",
					$parameters: {
						param1: "xstring",
						param2: "value in json"
					}
				},
				$linksFormParameters: {
					$title: "Form parameters",
					$description: "param1 and param2 from default, param3,param4 and param5 from form",
					$url: "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}&par3={param3}&par4={param4}&par4={param5}",
					$parameters: {
						"$properties": {
							param3: {
								$title: "mon param3",
								"$type": "application/x-string"
							},
							"param4": {
								"$title": "mon param4",
								"$type": "application/x-string"
							},
							"param5": {
								"$title": "mon param4",
								"$type": "application/x-string",
								"$isEditMode": false
							}
						},
						"param1": "xstring",
						"param2": "value"
					}
				}
			}

		},
		$actionsParameters: {
			$actions: {
				"actionEmployeesProperties": {
					"$title": "Action Employees Properties",
					"$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query",
					"$parameters": {
						"param1": "{lastName}",
						"param2": "value"
					}
				},
				"actionEmployeesPropertiesForm": {
					"$title": "Action Employees Properties Form",
					"$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query",
					"$parameters": {
						"$properties": {
							"param3": {
								"$title": "{@S-AOC0-125}",
								"$type": "application/x-string"
							},
							"param4": {
								"$title": "{@S-AOC0-8}",
								"$type": "application/x-string"
							},
							"param5": {
								"$title": "{@S-AOC1-1390}",
								"$type": "application/x-string",
								"$isEditMode": false
							}
						},
						"param1": "{lastName}",
						"param2": "value"
					}
				}
			}
		}

	}[value];
};
