// syracuse-ui/lib/tests/testPage.js
"use strict";
var helpers = require("syracuse-core/lib/helpers");
var DesktopPage = require("syracuse-ui/lib/page/desktopPage").DesktopPage;
var resources = require("./resources");

function TestPage() {}

exports.TestPage = helpers.defineClass(TestPage, DesktopPage, {
	onMenuClick: function(menuItem) {
		switch (menuItem.$bind) {
			case "$testUiLock":
				syra_site.uiLocker.lock(100, false);
				return false;
			case "$testUiLockLong":
				syra_site.uiLocker.lock(100, true);
				return false;
		}
		return true;
	},
	_initiliazeTestPage: function() {

	},
	getResources: function() {
		return resources;
	},
	onNavTestMenuClick: function(event, target) {
		syra_controller.executeMenu({
			$url: target.href
		}, this);
		event.preventDefault();
		event.stopPropagation();
	},
	loadBox: function(initData) {
		syra_site.pageLoader.initialize(this);
		this._initiliazeTestPage();
		this.$prototype.$properties = this.$prototype.$properties || {};
		DesktopPage.prototype.loadBox.call(this, initData);
		syra_site.uiTestController.addMenusToTestPage(this);
	}
});
