// syracuse-ui/lib/tests/testFactory.js
"use strict";
exports.diagnoses = require("./diagnosesTest").DiagnosesTest;
exports.messageBox = require("./messageBoxTest").MessageBoxTest;
exports.layouts = require("./layoutTest").LayoutTest;
exports.simpleFields = require("./simpleFieldsTest").SimpleFieldsTest;
exports.editorField = require("./editorFieldTest").EditorFieldTest;
exports.childFieldTest = require("./childFieldTest").ChildFieldTest;
exports.iconFields = require("./iconFieldsTest").IconFieldsTest;
exports.gridTest = require("./list/formatArrayTest").GridTest;
exports.scrollTest = require("./list/formatArrayTest").ScrollTest;
exports.vgridTest = require("./list/formatArrayTest").VgridTest;
exports.cardsTest = require("./list/formatArrayTest").CardsTest;
exports.editListTest = require("./list/editListTest").EditListTest;
exports.fillListTest = require("./list/formatArrayTest").FillListTest;
exports.selectionListTest = require("./list/formatArrayTest").SelectionListTest;
exports.graphTest = require("./list/graphTest").GraphTest;
exports.singleListTest = require("./list/singleListTest").SingleListTest;
exports.jsonPageViewer = require("./jsonPageViewer").JsonPageViewer;
exports.boostProcessTest = require("./boost/boostProcessTest").ProcessTest;
exports.messageBoxFk = require("./messageBoxTestFk").MessageBoxTest;
