// syracuse-ui/lib/tests/testManager.js
"use strict";

exports.robot = {
	logRequestCount: function(pendingRequests) {},
	logLock: function(pendingRequests) {},
	logPreferences: function() {},
	page_onAfterLoad: function(page) {}
};

exports.onAfterLogon = function(onAfterLogon) {
	if (syra_site.userProfile.dataset.enableTestRobot) {
		require.async('syracuse-ui/lib/tests/robotManager', function(err, module) {
			(exports.robot = module).onAfterLogon();
			onAfterLogon();
		});
	} else {
		onAfterLogon();
	}
};



function _openPage(segments) {
	syra_site.loadStyleSheet("uitest.css");
	require.async("syracuse-ui/lib/tests/pages/testPage", function(err, module) {
		try {
			var $itemPage = {
				layoutSlot: document.createElement("div"),
				$category: "page",
				openerUrlSegments: segments,
				$pageCategoryClass: module.TestPage,
				$representation: {
					$prototype: {
						$testName: segments.params.name.replace("_", "/"),
						$testCategory: segments.params.category,
						$properties: {},
						$links: {},
						$actions: {}
					},
					$article: {}
				}
			};
			$itemPage.openerUrlSegments.fullUrl = document.location.href;
			syra_site.onMainPageChange($itemPage);
		} catch (error) {
			console.error(error.message + "\n" + error.stack);
		}
	});
};


exports.controller_onChangeMainPage = function(segments) {
	if (segments && segments.$url && (segments.$url.indexOf("s-uitest") >= 0)) {
		_openPage(segments);
		return true;
	}
};

exports.page_onAfterLoad = function(page) {
	if (typeof window.callPhantom === 'function') {
		window.callPhantom({
			event: "onAfterLoad",
			autoFetch: true
		});
	}
	exports.robot.page_onAfterLoad(page);
};
