// syracuse-ui/lib/tests/uitestController.js
"use strict";
exports.$menus = [{
	$title: "DashboardConverter",
	$url: "?representation=s-uitest.$test&name=dashboardConverter"
}, {
	$title: "Array",
	$menus: [{
		$url: "?representation=s-uitest.$test&name=array_array&category=grid",
		$title: "Array Grid"
	}, {
		$url: "?representation=s-uitest.$test&name=array_array&category=gridcards",
		$title: "Array Grid - Cards"
	}, {
		$url: "?representation=s-uitest.$test&name=array_array&category=cards",
		$title: "Array Cards"
	}, {
		$url: "?representation=s-uitest.$test&name=array_array&category=selection",
		$title: "Array Selection"
	}, {
		$url: "?representation=s-uitest.$test&name=array_graph",
		$title: "Array Graph"
	}, {
		$url: "?representation=s-uitest.$test&name=array_tree",
		$title: "Array Tree "
	}, {
		$url: "?representation=s-uitest.$test&name=array_single",
		$title: "Array Single"
	}]

}, {
	$title: "Fields",
	$menus: [{
		$url: "?representation=s-uitest.$test&name=field_simple",
		$title: "Fields"
	}, {
		$url: "?representation=s-uitest.$test&name=field_binary",
		$title: "Fields Binary"
	}, {
		$url: "?representation=s-uitest.$test&name=field_editor",
		$title: "Fields Editor"
	}, {
		$url: "?representation=s-uitest.$test&name=field_child",
		$title: "Fields Child Object"
	}, {
		$url: "?representation=s-uitest.$test&name=field_icon",
		$title: "Fields Icon"
	}]
}, {
	$title: "Misc",
	$menus: [{
		$url: "?representation=s-uitest.$test&name=formula",
		$title: "Formula"
	}, {
		$url: "?representation=s-uitest.$$test&name=selectPrinter",
		$title: "Select Printer"
	}, {
		$url: "?representation=s-uitest.$test&name=process_process",
		$title: "Process"
	}, {
		$url: "?representation=s-uitest.$test&name=diagnoseCenter",
		$title: "Diagnoses"
	}, {
		$url: "?representation=s-uitest.$test&name=trackers_trackers",
		$title: "Trackers"
	}, {
		$url: "?representation=s-uitest.$test&name=parameters",
		$title: "Dynamic Parameters"
	}, {
		$url: "?representation=s-uitest.$test&name=macros_macros",
		$title: "Macros"
	}, {
		$url: "?representation=s-uitest.$test&name=layout",
		$title: "Layout"
	}]
}, {
	$title: "Charts",
	$menus: [{
		$url: "?representation=s-uitest.$test&name=charts_gantt",
		$title: "Gantt"
	}]

}, {
	$title: "Convergence",
	$menus: [{
		$url: "?representation=s-uitest.$test&name=convergenceCleaner",
		$title: "Convergence Cleaner",
	}, {
		$title: "X3 link --> AQFB : Test Tableaux",
		$url: "/trans/x3/erp/superv/$sessions?f=AQFB"
	}]
}, {
	$title: "Json",
	$menus: [{
		$title: "JsonLint",
		$url: "http://jsonlint.com/"
	}, {
		$title: "Jsonformatter Curiousconcept",
		$url: "http://jsonformatter.curiousconcept.com/"
	}]
}];


exports.openPage = function(openerUrlSegments) {
	syra_site.loadStyleSheet("uitest.css");
	syra_site.uiTestController = this;
	require.async("syracuse-ui/lib/tests/testPage", function(err, module) {
		try {
			var $itemPage = {
				layoutSlot: document.createElement("div"),
				$category: "page",
				openerUrlSegments: openerUrlSegments,
				$pageCategoryClass: module.TestPage,
				$representation: {
					$prototype: {
						$testName: openerUrlSegments.params.name.replace("_", "/"),
						$testCategory: openerUrlSegments.params.category,
						$properties: {},
						$links: {},
						$actions: {}
					},
					$article: {}
				}
			};
			$itemPage.openerUrlSegments.fullUrl = document.location.href;
			syra_site.onMainPageChange($itemPage);
		} catch (error) {
			console.error(error.message + "\n" + error.stack);
		}
	});

};
