// syracuse-ui/lib/tests/uitestController.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var _$menus = [{
	$title: "DashboardConverter",
	$url: "?representation=s-uitest.$test&screen=dashboardConverter"
}, {
	$title: "Array",
	$menus: [{
		$url: "?representation=s-uitest.$test&screen=array_grid",
		$title: "Grid"
	}, {
		$url: "?representation=s-uitest.$test&screen=array_edit",
		$title: "Edit"
	}, {
		$url: "?representation=s-uitest.$test&screen=array_fill",
		$title: "Fill"
	}, {
		$url: "?representation=s-uitest.$test&screen=array_cards",
		$title: "Cards"
	}, {
		$url: "?representation=s-uitest.$test&screen=array_gridcards",
		$title: "Grid - Cards"
	}, {
		$url: "?representation=s-uitest.$test&screen=array_selection",
		$title: "Selection"
	}, {
		$url: "?representation=s-uitest.$test&screen=array_tree",
		$title: "Tree view "
	}, {
		$url: "?representation=s-uitest.$test&screen=array_graph",
		$title: "Array Graph"
	}, {
		$url: "?representation=s-uitest.$test&screen=array_single",
		$title: "Array Single"
	}]

}, {
	$title: "Fields",
	$menus: [{
		$url: "?representation=s-uitest.$test&screen=field_simple",
		$title: "Fields"
	}, {
		$url: "?representation=s-uitest.$test&screen=field_binary",
		$title: "Binary Fields"
	}, {
		$url: "?representation=s-uitest.$test&screen=field_editor",
		$title: "Editor Field"
	}, {
		$url: "?representation=s-uitest.$test&screen=field_child",
		$title: "Child Object Fields"
	}, {
		$url: "?representation=s-uitest.$test&screen=field_icon",
		$title: "Icon Fields"
	}]
}, {
	$title: "Misc",
	$menus: [{
		$url: "?representation=s-uitest.$test&screen=formula",
		$title: "Formula"
	}, {
		$url: "?representation=s-uitest.$test&screen=selectPrinter",
		$title: "Select Printer"
	}, {
		$url: "?representation=s-uitest.$test&screen=process_process",
		$title: "Process"
	}, {
		$url: "?representation=s-uitest.$test&screen=diagnoseCenter",
		$title: "Diagnoses and Jobs"
	}, {
		$url: "?representation=s-uitest.$test&screen=parameters",
		$title: "Dynamic Parameters"
	}, {
		$url: "?representation=s-uitest.$test&screen=jsonviewer",
		$title: "Json"
	}, {
		$url: "?representation=s-uitest.$test&screen=macros",
		$title: "Macros"
	}]
}, {
	$title: "Charts",
	$menus: [{
		$url: "?representation=s-uitest.$test&screen=chart_bar",
		$title: "HighChart Bar"
	}, {
		$url: "?representation=s-uitest.$test&screen=chart_spider",
		$title: "HighChart Spider"
	}, {
		$url: "?representation=s-uitest.$test&screen=chart_gantt",
		$title: "Gantt"
	}]

}, {
	$title: "Layout",
	$menus: [{
		$url: "?representation=s-uitest.$test&screen=layout",
		$title: "Layout"
	}, {
		$url: "?representation=s-uitest.$test&screen=corestyle",
		$title: "Core Style"
	}]
}, {
	$title: "Convergence",
	$menus: [{
		$url: "?representation=s-uitest.$test&screen=convergenceCleaner",
		$title: "Convergence Cleaner",
	}, {
		$title: "X3 link --> AQFB : Test Tableaux",
		$url: "/trans/x3/erp/superv/$sessions?f=AQFB"
	}]
}, {
	$title: "Json",
	$menus: [{
		$title: "JsonLint",
		$url: "http://jsonlint.com/"
	}, {
		$title: "Jsonformatter Curiousconcept",
		$url: "http://jsonformatter.curiousconcept.com/"
	}]
}];

function _getTitle(screen) {
	for (var ii = 0, jj = _$menus.length; ii < jj; ii++) {
		var $main = _$menus[ii];
		if ($main.$menus) {
			for (var mm = 0, kk = $main.$menus.length; mm < kk; mm++) {
				var $menu = $main.$menus[mm];
				if ($menu.$url && $menu.$url.indexOf(screen) >= 0) {
					return $menu.$title;
				}
			}
		}
	}
	return screen;
}

exports.openPage = function(openerUrlSegments) {
	var screen = openerUrlSegments.params && openerUrlSegments.params.screen;
	var $links = {};

	syra_site.loadStyleSheet("uitest.css");
	require.async("syracuse-ui/lib/tests/" + screen.replace("_", "/"), function(err, module) {
		try {
			var $itemPage = {
				layoutSlot: document.createElement("div"),
				$category: "page",
				openerUrlSegments: openerUrlSegments,
				$pageCategoryClass: module.TestCase,
				$representation: {
					$prototype: {
						$properties: {},
						$title: _getTitle(screen),
						$links: $links,
						$actions: {
							$testUiLock: {
								"$title": "ui lock"
							},
							$testUiLockLong: {
								"$title": "ui lock long"
							}
						}
					},
					$article: {}
				}
			};
			$itemPage.openerUrlSegments.fullUrl = document.location.href;
			syra_site.onMainPageChange($itemPage);
		} catch (error) {
			console.error(error.message + "\n" + error.stack);
		}
	});

};
