// syracuse-ui/lib/authoring/fieldPalette/documentFieldPalette.js
"use strict";

exports.addProperties = function($properties, localize) {
	$properties.$documentField$isEmbedded = {
		$type: "application/x-boolean",
		$title: localize.fdoc_isEmbedded
	};
	$properties.$documentField$imageWidth = {
		$type: "application/x-string",
		$title: localize.fdoc_width
	};
	$properties.$documentField$imageHeight = {
		$type: "application/x-string",
		$title: localize.fdoc_height
	};
	$properties.$documentField$isAutoSize = {
		$type: "application/x-boolean",
		$title: localize.fdoc_isAutoHeight
	};
};

exports.getSection = function() {
	return {
		$category: "section",
		$authoringType: "documentField",
		$isTitleHidden: true,
		$layout: {
			$items: [{
				$bind: "$documentField$isAutoSize",
				$isEditMode: true,
				$skin: "s-aw-field"
			}, {
				$bind: "$documentField$isEmbedded",
				$isEditMode: true,
				$skin: "s-aw-field"
			}, {
				$bind: "$documentField$imageWidth",
				$isEditMode: true,
				$skin: "s-aw-field"
			}, {
				$bind: "$documentField$imageHeight",
				$isEditMode: true,
				$skin: "s-aw-field"
			}]
		}
	};
};

exports.onShow = function(awPalette, awItem) {
	if (awItem.$authoringType == "documentField") {
		awPalette.applyChange({
			$documentField$isAutoSize: awItem.$item.$isAutoSize,
			$documentField$imageWidth: awItem.$item.$imageWidth,
			$documentField$imageHeight: awItem.$item.$imageHeight,
			$documentField$isEmbedded: awItem.$item.$isEmbedded,
			$properties: {
				$documentField$isAutoSize: {
					isHidden: !awItem.isImage
				},
				$documentField$isEmbedded: {
					isHidden: awItem.isImage
				},
				$documentField$imageWidth: {
					isHidden: !awItem.isImage
				},
				$documentField$imageHeight: {
					isHidden: !awItem.isImage
				}
			}
		});
	}
};

exports.notifyDataChange = function(awPalette, awItem, metaData) {
	if (awItem.$authoringType == "documentField") {
		if (metaData.$documentField$isAutoSize != undefined ||
			metaData.$documentField$isEmbedded != undefined ||
			metaData.$documentField$imageHeight != undefined ||
			metaData.$documentField$imageWidth != undefined) {
			awItem.applyDesignMetaData({
				$isAutoSize: metaData.$documentField$isAutoSize,
				$imageWidth: metaData.$documentField$imageWidth,
				$imageHeight: metaData.$documentField$imageHeight,
				$isEmbedded: metaData.$documentField$isEmbedded
			}, true);
		}
	}
};
exports.applyChange = function(awPalette, awItem, newData) {
	if (newData.$documentField$isAutoSize !== undefined) {
		awPalette.applyChange({
			$properties: {
				$documentField$imageHeight: {
					$isDisabled: newData.$documentField$isAutoSize
				}
			}
		}, true);
	}
};
