// syracuse-ui/lib/office/siteLoader.js
"use strict";
var Site = require('syracuse-ui/lib/site/site').Site;

exports.load = function($officeSite, loader) {
	window.getStack = function() {
		try {
			var stack = [];
			var f = arguments.callee.caller;
			while (f) {
				var fn = f.toString();
				var fname = fn.substring(fn.indexOf('function') + 8, fn.indexOf('(')) || 'anonymous';
				stack.push(fname);
				f = f.caller;
			}
			return stack.join('\n');
		} catch (e) {}
	};
	window.onerror = function(message, url, line) {
		alert(message + '\nurl=' + url + '\nat line ' + line + '\n' + window.getStack());
		return false;
	};
	window.readBinaryURLContentIE = function(url, returnData) {
		try {
			var xhr = new XMLHttpRequest();
			var data = {};
			xhr.open('GET', url, false);

			xhr.setRequestHeader("X-Content-Type-Override", "text/plain; charset=x-user-defined");
			xhr.onreadystatechange = function(e) {};
			xhr.send();
			if (xhr.status == 200) {
				// this.responseText is broken in IE with binary data!
				// This is a c# byte array!
				returnData.setData(xhr.responseBody);
			} else
			if (xhr.status == 1223) {
				returnData.setErrorText("NOTFOUND");
			} else
			if (xhr.status == 404) {
				returnData.setErrorText("NOTFOUND");
			} else {
				returnData.setErrorText(xhr.response + "(" + xhr.status + ")");
			}
		} catch (e) {
			returnData.setErrorText(e);
		}
	};
	var site = new Site();
	site.$isSecurityTitleVisible = false;
	site.$officeSite = $officeSite;
	site.$isUserEndpointChangeDisabled = true;

	site.loadBox({
		$item: {
			$iconPath: "/syracuse-ui/themes/desktop/images/",
			$pageTemplateUrl: "/sdata/syracuse/collaboration/syracuse/pages('{endpoint}.{representation},{target},{pageview}')?role={$role}",
			$userProfileUrl: "/sdata/syracuse/collaboration/syracuse/userProfiles/$template/$workingCopies?representation=userProfile.$edit&role={$role}",
			$syracuseMainPageUrl: "/syracuse-main/html/main.html"
		},
		$prototype: {
			$links: {
				$home: {
					$title: "Home",
					$url: "?representation=" + $officeSite + ".$dashboard"
				}
			},
			$properties: {}
		}
	}, loader);
	return site;
};
