// syracuse-ui/lib/office/userProfile.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var RawPage = require("syracuse-ui/lib/page/rawPage").RawPage;
var date = require('syracuse-core/lib/types/date');
var _resources = date.resources;

function UserProfile() {}

exports.UserProfile = helpers.defineClass(UserProfile, RawPage, {
	getAuthoringLevel: function() {
		return "none";
	},
	dispose: function() {
		delete this.externalAdapter;
		delete this.formBlock;
		RawPage.prototype.dispose.call(this);
	},
	onMenuClick: function(menuItem) {
		if (menuItem.$bind == "$logout") {
			document.site.logout();
			return false;
		}
		return true;
	},
	ensureFormBlock: function() {
		var self = this;
		if (!self.formBlock) {
			self.formBlock = self.loadNewItem(self.$$item[0], {
				$category: "section",
				$title: document.site.localize.userProfile_up_welcome,
				$clientId: "s-user-form-block",
				$skin: "s-user-form",
				$layout: {
					$items: [{
						$bind: "$save",
						$category: "link",
						$isHidden: true,
						onServerRequest: function(menu) {
							//this is item menu	
							if (menu.$links && menu.$links.$location) {
								var currentPage = menu.boxParent.getArticle();
								self.validatePrototype(self.$prototype);
								var location = ((document.site.$prototype || {}).$links || {}).$home;
								location && document.controller.openPage(self, location);
							}
							return false;
						}
					}, {
						$isEditMode: true,
						$isTopLabelAlignment: false,
						$isReferenceTitleVisible: false,
						$skin: "s-user-profile-ref",
						$css: "s-user-profile-selectedRole",
						$bind: "photo"
					}, {
						$isEditMode: true,
						$isTopLabelAlignment: false,
						$isReferenceTitleVisible: false,
						$skin: "s-user-profile-ref",
						$css: "s-user-profile-selectedRole",
						$bind: "selectedRole",
						$title: document.site.localize.userProfile_up_role
					}, {
						$isEditMode: true,
						$isTopLabelAlignment: false,
						$skin: "s-user-profile-ref",
						$isReferenceTitleVisible: false,
						$css: "s-user-profile-selectedLocale",
						$bind: "selectedLocale",
						$title: document.site.localize.userProfile_up_locale
					}]
				}
			});
			self.isPageLoaded = true;
		}
	},
	drawBox: function() {
		this.externalAdapter = document.site.externalAdapter;
		this.externalAdapter = document.site.externalAdapter;
		this.domItem = document.createElement("div");
		this.domItem.id = "s-user-profile";
		this.$$item = $(this.domItem);
	},
	showDiagnoses: function(message) {
		//afficher erreur ds popup ???
	},
	notifyDataChange: function(field, value) {
		this.notifyActionChange(null, this, {
			$save: {
				$isRequested: true
			}
		}, false);
		RawPage.prototype.notifyDataChange.call(this, field, value);
	},
	saveSitePreferences: function() {},
	applyChange: function(newData) {
		// refresh page when user changes locale
		var sl = this.getSelectedLocale();
		if (Object.keys(this.dataset).length > 0) {
			if (sl && newData.selectedLocale && JSON.stringify(sl) != JSON.stringify(newData.selectedLocale))
				location.reload();
		}

		if (newData && newData.selectedRole) {
			document.site.$prototype.$role = newData.selectedRole.$uuid;
		}

		RawPage.prototype.applyChange.call(this, newData);
		document.site.onUserProfileChange();
	},
	showPageSecurity: function(page) {
		if (document.site.$item.$isPageSecurityView !== false) {
			if (this.dataset && page.showPageSecurity) {
				page.showPageSecurity(this.dataset.selectedRole && this.dataset.selectedRole.description, this.dataset.selectedEndpoint && this.dataset.selectedEndpoint.description);
			}
		}
	},
	getDataset: function() {
		return (this.dataset = this.dataset || {});
	},
	getSelectedLocale: function() {
		return this.getDataset().selectedLocale || {};
	},
	getSelectedEndpointString: function() {
		/*        var endpoint = this.dataset ? this.dataset.selectedEndpoint : null;
         return endpoint ? (endpoint.application + "." + endpoint.contract + "." + endpoint.dataset) : null;
         */
		return null;
	},
	getDateFormat: function(displayFormat) {
		var selectedLocale = this.getSelectedLocale();
		return displayFormat == "DD" ? (selectedLocale.longDate || "dd MMMM yyyy") : (selectedLocale.shortDate || "yyyy-MM-dd");
	},
	getTimeFormat: function(displayFormat) {
		var selectedLocale = this.getSelectedLocale();
		return displayFormat == "TT" ? (selectedLocale.longTime || "HH:mm:ss") : (selectedLocale.shortTime || "HH:mm");
	},
	getDatetimeFormat: function(displayFormat) {
		var selectedLocale = this.getSelectedLocale();
		var format;
		switch (displayFormat) {
			case "F":
				// sortable datetime format
				format = _resources().formatPatterns ? _resources().formatPatterns["sortableDateTime"] : "yyyy-MM-ddTHH:mm:ss";
				break;
			case "FF":
				// full datetime format
				format = selectedLocale.selectedLocale.longDatetime || (_resources().formatPatterns ? _resources().formatPatterns["fullDateTime"] : "dd MMMM yyyy HH:mm:ss");
				break;
			case "U":
				// universal sortable datetime format
				format = _resources().formatPatterns ? _resources().formatPatterns["universalSortableDateTime"] : "yyyy-MM-dd HH:mm:ssZ";
				break;
			default:
				// datetime default format is dateformat + space + timeformat
				format = selectedLocale.shortDatetime || (this.getDateFormat() + " " + this.getTimeFormat());
		}
		return format;
	},
	getNumberFormatObj: function(type) {
		var selectedLocale = this.getSelectedLocale();
		var f = "#,";
		var numberGroupSize = selectedLocale.numberGroupSize || "3";
		for (var i = 0; i < numberGroupSize; i++) {
			f += (i == numberGroupSize - 1) ? "0" : "#";
		}
		f += type == "application/x-integer" ? '' : '.##';
		// numFormat default -> #,##0 (integer) or #,##0.## (decimal)
		return {
			numFormat: f,
			decimalSeparator: selectedLocale.numberDecimalSeparator || ".",
			groupSeparator: selectedLocale.numberGroupSeparator || ",",
			groupSize: numberGroupSize
		};
	},
	getNumberFormat: function(type) {
		return this.getNumberFormatObj(type).numFormat;
	},
	getTwoDigitYearMax: function() {
		return this.getSelectedLocale().twoDigitYearMax || 2029;
	},
	getCurrencyDisplay: function() {
		return this.getDataset().currencyDisplay || "symbol";
	},
	getUnitDisplay: function() {
		return this.getDataset().unitDisplay || "symbol";
	},
	getCurrencyPosition: function() {
		return this.getDataset().currencyPosition || "right";
	},
	buildWCContent: function() {
		var dd = this.dataset || {};
		return ["selectedEndpoint", "selectedRole", "selectedLocale"].reduce(function(prev, item) {
			var val = dd[item];
			prev[item] = (val && val.$uuid) ? {
				$uuid: val.$uuid
			} : null;
			return prev;
		}, {
			$etag: dd.$etag
		});
	}
});
