// @sage/syracuse-react/src/style/styleSheet.ts
"use strict";
import * as Sdata from '../dataStore/sdata';

export const PATH = "/syracuse-main/html/styles/";
export const DEFAULT_THEME_UUID = "-1";

let styleSheets:any = {};

let themeStyleSheet:HTMLLinkElement;

export function unload(name:string) {
	let link = styleSheets[name];
	link && document.head.removeChild(link);
	delete styleSheets[name];
}

export function load(name :string, isExtended?:boolean) {
	if (!isExtended) {
		name =  PATH + name;
	}
	if (!styleSheets[name]) {
		var link = styleSheets[name] = document.createElement("link");
		link.setAttribute("rel", "stylesheet");
		link.setAttribute("type", "text/css");
		link.setAttribute("data-s-name", name);
		link.href =  name;
		if (themeStyleSheet) {
			document.head.insertBefore(link, themeStyleSheet);
		} else {
			document.head.appendChild(link);
		}
		return link;
	}
}

export interface ITheme  extends Sdata.IEntity{
	description: string
	cssFiles: Array<{
		path:string
	}>
}

export function applyTheme(theme :ITheme) {
	let current = themeStyleSheet && themeStyleSheet.getAttribute("data-s-name") 
	if (theme && theme.cssFiles) {
		let name = theme.cssFiles[0].path.toLowerCase();
		if (name.indexOf(".css") < 0) {
			name = "custom/" + name + ".css";
			if (themeStyleSheet) {
				if (current == name) {
					return;
				}
				unload(current);
			}
			themeStyleSheet = null;
			themeStyleSheet = load(name);
		}
	} else {
		current && unload(current);
		themeStyleSheet = null;
	}
}
