// @sage/syracuse-ui/lib/field/vignette/pageVignetteField.js
"use strict";
var _helpers = require('@sage/syracuse-core').helpers;
var VignetteField = require("syracuse-ui/lib/field/vignette/vignetteField").VignetteField;

function _loadContentPage(field) {
	if (!field._isContentPageRequested) {
		field._isContentPageRequested = true;
		if (field.$url || !syra_fields.embedded.hasMasterChildRelation(field)) {
			//if masterChild, vignette is lazy loaded
			field.loadContentRepresentation();
		}
	}
}

function PageVignetteField() {}

exports.PageVignetteField = _helpers.defineClass(PageVignetteField, VignetteField, {
	addResizer: function() { //temp disable of resizer
	},
	onAfterMaximize: function() {
		this.vignette && this.vignette.resizeItem(true);
	},
	loadContentRepresentation: function() {
		syra_router.loadVignette(this);
	},
	designArticle: function(open) {
		if (this.vignette) {
			switch (this.vignette.$facet) {
				case "$query":
				case "$cube":
					var fields = this.vignette.layoutContent.getFields();
					for (var ii = 0, jj = fields.length; ii < jj; ii++) {
						if (fields[ii].isArrayField) {
							var open = !fields[ii].designer;
							fields[ii].designArticle(open);
							break;
						}
					}
					break;
			}
		}
	},
	renderLayoutContent: function($url, onVignetteRefresh) {
		VignetteField.prototype.renderLayoutContent.call(this);
		if ($url) {
			this.$url = $url;
		} else {
			if (!(this.$url && this.$masterBinds)) {
				this.$url = this.$field.$altLocation && this.$field.$altLocation.$url;
				if (this.$url && this.page.ldpRecord) {
					this.$url = syra_expression.parse(this.page.ldpRecord.page, this.$url);
				}
			}
		}
		this._isContentPageRequested = false;
		this._onVignetteRefresh = onVignetteRefresh;
		_loadContentPage(this);
	},
	resizeItem: function(doResize) {
		_loadContentPage(this);
		this.vignette && this.vignette.resizeItem(doResize);
	},
	dispose: function() {
		syra_fields.embedded.disposeMasterChildRelation(this);
		this.syraVignettePage && this.syraVignettePage.dispose();
		this.vignette && this.vignette.dispose();
		VignetteField.prototype.dispose.call(this);
	}
});
