// @sage/syracuse-ui/lib/field/vignette/vignetteField.js
"use strict";
var helpers = require('@sage/syracuse-core').helpers;
var Section = require("@sage/syracuse-ui/lib/item/section").Section;
var Resizer = require("@sage/syracuse-ui/lib/scroll/resizer").Resizer;

function VignetteField() {}

exports.VignetteField = helpers.defineClass(VignetteField, Section, {
	ensureSkin: function() {
		Section.prototype.ensureSkin.call(this);
		this.$designLevel = "field";
	},
	resizeItem: function() {
		if (this.$item.$height) {
			this.body.style.height = this.$item.$height + "px";
		}
	},
	setValue: function() {},
	load: function() {
		this.isVignetteField = true;
		if (!this.page.isLandingPage && this.$item.$isBoxCollapsable == undefined) {
			this.$item.$isBoxCollapsable = true;
		}
		Section.prototype.load.call(this);
		//$(this.domItem).resizable();
		this.$pagePreferences = syra_preference.page.getField(this);
		if (this.$pagePreferences.$height) {
			this.$item.$height = this.$pagePreferences.$height;
		}
		this.addRefreshButton();
		syra_maximizer.addButton({
			parent: this,
			slot: this.ensureButtonsSlot(),
			css: this.$skin + "-btn"
		});
		syra_item.bind(this, this.$item.$bind);

		this.page.resizableList.add(this);
		this.addResizer();
	},
	addRefreshButton: function() {
		this.refreshBtn = syra_button.add({
			parent: this,
			slot: this.ensureButtonsSlot(),
			text: syra_local.box_refresh,
			css: this.$skin + "-btn",
			iconOnly: true,
			fontIcon: "refresh",
			click: function() {
				this.parent.refresh(true);
			}
		});
	},
	refresh: function(onVignetteRefresh) {
		var $refresh = this.vignette && this.vignette.$menus && this.vignette.$menus.$refresh;
		var $url = $refresh && $refresh.$url;
		if ($url) {
			$url = syra_expression.parse(this.vignette, $url);
		}
		this.renderLayoutContent($url, onVignetteRefresh);
	},
	hideHeader: function(hide) {
		syra_dom.hide(this.domTitle, hide);
		this.header.className = hide ? "s-vignette-field-header" : (this.$skin + "-head");
	},
	onItemInOut: function(onEnter, event) {},
	toggleMenuButton: function(show) {
		if (show) {
			if (!this.menuBtn) {
				this.menuBtn = syra_button.add({
					parent: this,
					slot: this.ensureButtonsSlot(),
					text: syra_local.box_menus,
					css: this.$skin + "-btn",
					iconOnly: true,
					fontIcon: "menus",
					click: function() {
						syra_menus.open({
							scope: this.parent.vignette,
							picker: this
						});
					}
				});
			}
		} else {
			syra_button.remove(this.menuBtn);
		}
	},
	renderLayoutContent: function() {
		if (this.vignette) {
			syra_dom.empty(this.body);
			this.vignette.dispose();
			this.vignette = null;
		}
	},
	addResizer: function() {
		this.resizer = new Resizer({
			item: this,
			body: this.body,
			slot: this.domItem,
			handles: ["s"],
			start: function(resizing) {
				if (this._iframe) {
					this._iframe.style.pointerEvents = "none";
				}
			},
			drag: function(resizing) {

			},
			stop: function(resizing) {
				if (this._iframe) {
					this._iframe.style.pointerEvents = "auto";
				}
				var field = this.item;
				field.$pagePreferences.$height = field.$item.$height = this.$height;
				syra_preference.page.saveField(field, field.$pagePreferences);
				if (field.page.designer) {
					field.page.designer.endArticleUpdate();
				} else {
					field.page.savePageDesign();
				}
			}
		});
	},
	dispose: function() {
		this.resizer && this.resizer.dispose();
		this.toggleMenuButton(false);
		this.page && this.page.resizableList.remove(this);
		Section.prototype.dispose.call(this);
	}
});
