// syracuse-ui/lib/alert/alertField.js
"use strict";


function _getFieldItems(panel, field) {
	var items = [];
	for (var ii = 0, jj = panel.fieldStack.length; ii < jj; ii++) {
		var line = panel.fieldStack[ii];
		if (line.field && line.field == field) {
			items.push(line);
		}
	}
	return items;
}

function _addMessage(field, $diagnose) {
	if ($diagnose.$message) {
		syra_button.add({
			isIndicator: true,
			parent: field,
			slot: field.diagsSlot,
			text: $diagnose.$message,
			css: "s-diag-" + $diagnose.$severity + " s-diag-field",
			fontIcon: $diagnose.$severity == "info" ? $diagnose.$severity : "diagnose"
		});
	}
}

function _refreshParentBoxes(box, update, add) {
	while (box) {
		box.severitiesCount = box.severitiesCount || {};
		var prioritySeverity;
		var severities = ["fatal", "error", "warning", "info", "success"];
		for (var ii = 0, jj = severities.length; ii < jj; ii++) {
			var severity = severities[ii];
			if (update[severity]) {
				if (add) {
					box.severitiesCount[severity] = (box.severitiesCount[severity] || 0) + update[severity];
				} else {
					box.severitiesCount[severity] -= update.fatal;
				}
			}
			if (box.severitiesCount[severity]) {
				if (!prioritySeverity) {
					prioritySeverity = severity;
				}
			}
		}
		if (box.isTabSection || (box.header && box.$item.$isBoxCollapsable)) {
			if (add && box.prioritySeverity != prioritySeverity) {
				box.diagnoseBtn = syra_button.add({
					parent: box,
					slot: box.ensureButtonsSlot(),
					text: syra_local.box_diagnoses,
					css: box.$skin + "-btn",
					iconOnly: true,
					fontIcon: prioritySeverity == "info" ? prioritySeverity : "diagnose"
				});
				box.diagnoseBtn.link.className += " s-diag-" + prioritySeverity;
				box.diagnoseBtn.link.title = "";
				box.diagnoseBtn.link.syrainout = box.id;
			}
			box.prioritySeverity = prioritySeverity;
			if (!box.prioritySeverity) {
				syra_button.remove(box.diagnoseBtn);
				delete box.diagnoseBtn;
			}
		}
		box = box.boxParent;
	}
}

function _onClickGoToField() {
	var field;
	if (this.fieldArticle) {
		field = this.fieldArticle.boundFields[this.$fieldBind];
		if (!field && this.fieldArticle.rowCardBtn) {
			this.fieldArticle.rowCardBtn.click();
			field = this.fieldArticle.boundFields[this.$fieldBind];
		}
		field = field && field[0];
	} else {
		field = syra_item.get(this.fieldId);
	}
	if (field) {
		field.focus && field.focus();
		exports.walkToField(this.parent, 0, field);
	}
}

function _cleanUnloadedFieldMessage(panel, field) {
	var cleaned = [];
	for (var ii = 0, jj = panel.fieldStack.length; ii < jj; ii++) {
		var line = panel.fieldStack[ii];
		if (line.$field && line.article == field.articleParent) {
			cleaned.push(line);
		}
	}
	for (var ii = 0, jj = cleaned.length; ii < jj; ii++) {
		panel.fieldStack.splice(panel.fieldStack.indexOf(cleaned[ii]), 1);
		line.clean();
	}
}

function _showFieldMessage(panel, message) {
	var field = message.field;
	field.isDiagnosePopup = field.$item.$inplace && !field.$item.$isDiagnoseInline;
	_cleanUnloadedFieldMessage(panel, field);
	if (!field.isDiagnoseDisabled && message.$diagnoses && message.$diagnoses.length > 0) {
		if (field.isDiagnosePopup) {
			if (!field.diagnoseFlag && !syra_site.isTabletDevice) {
				var severity = message.$diagnoses[0].$severity;
				field.diagnoseFlag = syra_button.add({
					isIndicator: true,
					parent: field,
					slot: field._dataValue,
					iconOnly: true,
					css: "s-diag-" + severity + (field.$isEditMode ? " s-diag-inline-edit" : " s-diag-inline"),
					fontIcon: severity == "info" ? severity : "diagnose"
				});
			}
		} else {
			field.diagsSlot = document.createElement("div");
			field.diagsSlot.className = "s-diag-field-slot";
			syra_dom.toggleClass(field.refDescriptionItem || field.descriptionItem, "s-field-desc-diagnose", true);
		}
		var severities = {};
		for (var ii = 0, jj = message.$diagnoses.length; ii < jj; ii++) {
			var line = panel.addLine(field.page, message.$diagnoses[ii]);
			line.article = field.articleParent;
			line.field = field;
			panel.fieldStack.push(line);

			if (field.isDiagnosePopup) {
				var text = "";
				if (field.articleParent.getRecordIndex) {
					text = " (line " + (field.articleParent.getRecordIndex() + 1) + " )";
				}
				line.fieldText = (field.getTitle() || field.$field.$title || "") + text;
			} else {
				// title for single record collections if not in field data
				if (!field.getTitle() && !field.$field.$title && field.arrayLevel && field.arrayLevel == "cell") {
					var parent = field.articleParent.articleParent;
					var title = parent.titleLabel ? parent.titleLabel : (parent.$field ? parent.$field.$title : title);
					line.fieldText = title + ":";
				} else {
					line.fieldText = (field.getTitle() || field.$field.$title || field.$item.$bind) + ":";
				}
				_addMessage(field, line.$diagnose);
			}

			var gotoBtn = syra_button.add({
				parent: panel,
				slot: line.link,
				text: line.fieldText,
				css: "s-diagnose-msg-link",
				click: _onClickGoToField,
				fieldId: field.id
			});
			line.addTopLabel();
			severities[line.$diagnose.$severity] = (severities[line.$diagnose.$severity] || 0) + 1;
			if (field.$item.$isExpressionChild) {
				var dom = document.createElement("span");
				dom.setAttribute("data-s-menu", field.id);
				dom.textContent = gotoBtn.link.textContent.toUpperCase() + " ";
				line.msgItem.link.insertBefore(gotoBtn.link, line.msgItem.iconflag.nextSibling);
				if (field.diagsSlot) {
					syra_dom.hide(field.diagsSlot, true);
				}
			} else {
				line.msgItem.link.insertBefore(gotoBtn.link, line.msgItem.iconflag.nextSibling);
			}
		}
		_refreshParentBoxes(field.boxParent, severities, true);
		syra_dom.toggleClass(field._dataValue, "s-" + line.$diagnose.$severity, true);
		field.diagsSlot && field._core.appendChild(field.diagsSlot);
		field.$setValue && field.$setValue();
	}
}

function _showUnloadedFieldMessage(panel, message) {
	var article = message.article;
	var $field = article.$prototype.$properties[message.$bind];
	for (var ii = 0, jj = message.$diagnoses.length; ii < jj; ii++) {
		var line = panel.addLine(article.page, message.$diagnoses[ii]);
		line.article = article;
		line.$field = $field;
		panel.fieldStack.push(line)
		var text = "";
		if (line.article.getRecordIndex) {
			text = " (line " + (line.article.getRecordIndex() + 1) + " )";
		}
		line.fieldText = syra_expression.parse(line.article, line.$field.$title || "") + text;
		var gotoBtn = syra_button.add({
			parent: panel,
			slot: line.link,
			text: line.fieldText,
			css: "s-diagnose-msg-link",
			click: _onClickGoToField,
			$fieldBind: message.$bind,
			fieldArticle: line.article
		});
		line.addTopLabel();
		line.msgItem.link.insertBefore(gotoBtn.link, line.msgItem.iconflag.nextSibling);
	}
}

exports.show = function(panel, message) {
	if (message.field) {
		_showFieldMessage(panel, message);
	} else {
		_showUnloadedFieldMessage(panel, message);
	}
};

exports.clean = function(panel, field) {
	if (field.diagsSlot) {
		syra_dom.toggleClass(field.refDescriptionItem || field.descriptionItem, "s-field-desc-diagnose", false);
		syra_dom.remove(field.diagsSlot);
		delete field.diagsSlot;
	}
	syra_button.remove(field.diagnoseFlag);
	delete field.diagnoseFlag;

	var css = field._dataValue.className;
	var severities = {};
	var items = _getFieldItems(panel, field);
	for (var ii = 0, jj = items.length; ii < jj; ii++) {
		var item = items[ii];
		severities[item.$diagnose.$severity] = (severities[item.$diagnose.$severity] || 0) + 1;
		css = css.replace("s-" + item.$diagnose.$severity, "");
		item.clean();
	}
	field._dataValue.className = css;
	_refreshParentBoxes(field.boxParent, severities, false);
};

exports.onFieldItemInOut = function(panel, field, onEnter) {
	var items = _getFieldItems(panel, field);
	if (field.isDiagnosePopup && items) {
		if (onEnter) {
			if (!field.diagsSlot) {
				if (items.length) {
					field.diagsSlot = syra_dom.div("s-diag-field-slot-popup");
					for (var ii = 0, jj = items.length; ii < jj; ii++) {
						_addMessage(field, items[ii].$diagnose);
					}
					field.diagsPopup = syra_over.openPopup(field.page, {
						slot: field.diagsSlot,
						position: {
							my: "left top",
							at: "left bottom",
							of: field.domItem
						},
						close: function() {
							syra_dom.remove(field.diagsSlot);
							field.diagsSlot = field.diagsPopup = null;
						}
					});
				}
			}
		} else {
			field.diagsPopup && field.diagsPopup.close();
		}
	}
};


exports.walkToField = function(panel, btn, field) {
	if (panel.fieldStack && panel.fieldStack.length > 0) {
		var focused = panel._focused;
		var items = field && _getFieldItems(panel, field);
		if (!btn && items && items.length) {
			focused = items[0];
		} else {
			var first, last, next;
			for (var ii = 0, jj = panel.fieldStack.length; ii < jj; ii++) {
				var last = panel.fieldStack[ii];
				if (first == undefined) {
					first = last;
				}
				if (panel._focused && panel._focused.field == last.field) {
					next = true;
				} else {
					if (next === true) {
						next = last;
					}
				}
			}
			if (panel._focused) {
				focused = panel._focused == last ? first : ((next === true) ? first : next);
			} else {
				focused = first;
			}!focused.field.disposed && focused.field.focus && focused.field.focus();
		}
		focused.focus();
		panel.refresh();
	}
};
