// syracuse-ui/lib/alert/alertDialog.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var DesktopPage = require("syracuse-ui/lib/page/desktopPage").DesktopPage;

function _preformat(txt) {
	return "<pre class='s-msgbox-pre'>" + syra_dom.formatHTMLText(txt, true) + "</pre>";
}

function _detailToString(obj) {
	if (obj == undefined)
		return "";
	try {
		if (typeof(obj) == 'string') {
			if (obj.length == 0)
				return obj;
			obj = JSON.parse(obj);
		}
		return JSON.stringify(obj, null, "   ");
	} catch (e) {
		return obj.toString();
	}
}

function _addDiagnoses(page, $diagnoses) {
	_ensureStackTrace(page);
	// add to formatted tab content
	var diagnoses = syra_dom.addDiv("s-msgbox-diagnoses", page._coreFormatted);

	page._groups = {};
	var severities = ["error", "fatal", "warning", "info"];
	for (var ii = 0, jj = severities.length; ii < jj; ii++) {
		var severity = severities[ii];
		page._groups[severity] = {
			groupItem: syra_dom.addDiv('s-msgbox-group-' + severity),
			messages: []
		};
	}
	for (var ii = 0, jj = $diagnoses.length; ii < jj; ii++) {
		var $diagnose = $diagnoses[ii];
		var severity = $diagnose.severity || $diagnose.$severity;
		var group = null;
		if (severity) {
			severity = severity.toLowerCase();
			group = page._groups[severity];
		}
		if (!group) {
			group = page._groups[severity = "error"];
		}
		var _code = $diagnose.appCode != null && $diagnose.appCode != undefined ? $diagnose.appCode : ($diagnose.sdataCode != null && $diagnose.sdataCode != undefined ? $diagnose.sdataCode : null);
		var message = {
			label: document.createElement("div"),
			stackTrace: $diagnose.stackTrace || $diagnose.$stackTrace,
			appCode: _code != null ? (syra_local.msgbox_diag_appCode + _code) : undefined
		};
		group.messages.push(message);
		message.label.className = "s-msgbox-label";
		message.label.innerHTML = syra_dom.formatHTMLText($diagnose.message || $diagnose.$message || syra_local.msgbox_diagnose_default_msg, true);
		if ($diagnose.stackTrace || $diagnose.$stackTrace) {
			syra_button.add({
				parent: page,
				slot: message.label,
				text: syra_local.msgbox_diagnose_show_detail,
				iconOnly: true,
				css: "s-msgbox-diag-btn",
				fontIcon: "expand_l",
				message: message,
				click: _onMessageClick
			});
		}
		var flag = document.createElement("em");
		flag.className = "s-msgbox-label-" + severity;
		flag.textContent = ".";
		message.label.insertBefore(flag, message.label.firstChild);
		group.groupItem.appendChild(message.label);
		if ($diagnose.$links || $diagnose.$actions) {
			page.applyChange($diagnose);
		}
		diagnoses.appendChild(group.groupItem);
	}
	// add to raw tab content
	page._coreRaw.innerHTML += _preformat(_detailToString($diagnoses));
}

function _ensureStackTrace(page) {
	if (!page._core) {
		var footer = page.idMap["$footer"];
		page._coreTabsNav = document.createElement("nav");
		page._coreTabsNav.className = "s-msgbox-core-tabs-nav";
		page._coreTabsNav.style.display = "none";
		footer.domItem.parentNode.insertBefore(page._coreTabsNav, footer.domItem);

		page._core = document.createElement("div");
		page._core.className = "s-msgbox-core";
		page._core.style.display = "none";
		footer.domItem.parentNode.insertBefore(page._core, footer.domItem);

		// tabs
		page._openedTab = syra_button.add({
			parent: page,
			slot: page._coreTabsNav,
			text: syra_local.msgbox_nav_title_formatted,
			css: "s-msgbox-core-tab s-open",
			click: _onTabClick
		});
		syra_button.add({
			parent: page,
			slot: page._coreTabsNav,
			text: syra_local.msgbox_nav_title_raw,
			css: "s-msgbox-core-tab",
			click: _onTabClick
		});

		// tabs content
		page._coreFormatted = syra_dom.addDiv("s-msgbox-core-tab-formatted", page._core);

		page._coreRaw = syra_dom.addDiv("s-msgbox-core-tab-raw");
		page._coreRaw.style.display = "none";
		page._core.appendChild(page._coreRaw);

		var slot = syra_dom.addDiv("s-msgbox-btn-slot");
		page._bodyBtn = syra_button.add({
			parent: page,
			slot: slot,
			text: syra_local.msgbox_body_picker_closed,
			css: "s-msg-box-body-btn",
			fontIcon: "expand_s",
			click: function() {
				var msgBox = this.parent;
				syra_dom.toggle(msgBox._coreTabsNav);
				syra_dom.toggle(msgBox._core);
				this.isOpened = !this.isOpened;
				syra_button.setText(this, this.isOpened ? syra_local.msgbox_body_picker_opened : syra_local.msgbox_body_picker_closed, this.isOpened ? "collapse_s" : "expand_s");
				syra_button.hide(msgBox._selectBtn, !this.isOpened);
				msgBox.resizeItem();
			}
		});
		page._message.parentNode.insertBefore(slot, page._message);

		slot = syra_dom.addDiv("s-msgbox-btn-slot");
		page._selectBtn = syra_button.add({
			parent: page,
			slot: slot,
			text: syra_local.msgbox_select_picker,
			css: "s-msgbox-select-btn",
			isHidden: true,
			click: function() {
				this.parent.selectText();
			}
		});
		page._message.parentNode.insertBefore(slot, page._message);
	}
}

function _addAutoClose(page) {
	var count = (typeof(page.$item.$isAutoClose) == "number" ? page.$item.$isAutoClose : 4000) / 1000;
	page.domTitle.textContent = page.$item.$title + "  (" + count + ")";
	page.autoCloseTimeOut = setInterval(function() {
		var countTxt = "(" + (--count) + ")";
		page.domTitle.textContent = page.$item.$title + " " + countTxt;
		if (count == 0) {
			clearInterval(page.autoCloseTimeOut);
			page.close("auto");
		}
	}, 1000);
}

function _onTabClick() {
	var msgBox = this.parent;
	msgBox._openedTab.link.className = "s-msgbox-core-tab";
	(msgBox._openedTab = this).link.className = "s-msgbox-core-tab s-open";
	syra_dom.toggle(msgBox._coreFormatted);
	syra_dom.toggle(msgBox._coreRaw);
	msgBox.resizeItem();
}

function _onMessageClick() {
	var page = this.parent;
	this.isOpened = !this.isOpened;
	syra_button.setText(this, this.isOpened ? syra_local.msgbox_diagnose_hide_detail : syra_local.msgbox_diagnose_show_detail, this.isOpened ? "collapse_l" : "expand_l");
	var message = this.message;
	if (this.isOpened) {
		if (!this.message.stackTraceView) {
			this.message.stackTraceView = syra_dom.addDiv("s-msgbox-diag-stackTrace", this.message.label);
			this.message.stackTraceView.innerHTML = _preformat(((this.message.appCode != undefined ? this.message.appCode + "\n" : "") + (this.message.stackTrace || this.message.$stackTrace)).replace(/\n/g, "<br/>"));
		}
	}
	syra_dom.hide(this.message.stackTraceView, !this.isOpened);
	page.resizeItem();
}

function AlertDialog() {}

exports.AlertDialog = helpers.defineClass(AlertDialog, DesktopPage, {
	loadBox: function(initData) {
		this.$item.$skin = "s-msgbox";
		this.isFooterDisabled = true;
		this.isBreadCrumbHidden = true;
		this.isSecurityViewHidden = true;
		if (!this.$item.$buttons) {
			switch (this.$item.$type) {
				case "question":
					this.$item.$buttons = "yesno";
					break;
				case "prompt":
					this.$item.$buttons = "okcancel";
					break;
				default:
					this.$item.$buttons = "ok";
					break;
			}
		}

		this.isAlertDialog = true;
		this.$prototype.$links = {};
		this.$itemButtons = [];
		var ids = ["yes", "no", "ok", "cancel"];
		for (var ii = 0, jj = ids.length; ii < jj; ii++) {
			var id = ids[ii];
			if (this.$item.$buttons.indexOf(id) >= 0) {
				this.$prototype.$links[id] = {
					$title: this.$item.$buttonsTitle && this.$item.$buttonsTitle[id] || syra_local["msgbox_" + id]
				};
				this.$itemButtons.push({
					$bind: id
				});
				if ((this.$item.$default || "ok") == id) {
					this._defaultBtnId = id;
				}
			}
		}
		if (!this._defaultBtnId) {
			this._defaultBtnId = this.$itemButtons[0].$bind;
		}

		this.$prototype.$properties = {
			$message: {
				$type: "application/x-string",
				$title: "$isTitleHidden",
			}
		};
		this.$item.$layout = {
			$items: [{
				$bind: "$message",
				$isTitleHidden: true,
				$skin: this.$item.$skin + "-msg"
			}]
		};
		if (this.$item.$type == "prompt") {
			this.$prototype.$properties.$prompt = {
				$type: "application/x-string"
			};
			this.$item.$layout.$items.push({
				$bind: "$prompt",
				$isEditMode: true,
				$isAutoSizeDisabled: true,
				$maxLength: 50,
				$isTitleHidden: true,
				$skin: this.$item.$skin + "-prompt"
			});
		}
		this.$item.$layout.$items.push({
			$skin: "s-msgbox-foot",
			$clientId: "$footer",
			$category: "menus",
			$layout: {
				$items: [{
					$category: "menus",
					$skin: "s-msgbox-links",
					$isMenusBag: true,
					$isHidden: this.$item.$useLinkAsOk
				}, {
					$skin: "s-msgbox-buttons",
					$layout: {
						$layoutType: "row",
						$items: this.$itemButtons
					}
				}]
			}
		});

		DesktopPage.prototype.loadBox.call(this, initData);

		if (!this.$item.$displayDisabled) {
			if (this.$item.layoutSlot) {
				(this.layoutSlot = this.$item.layoutSlot).appendChild(this.domItem);
				this.domItem.style.display = "inline-block";
			} else {
				this.domItem.style.position = "fixed";
				this.dialogWrapper = syra_dlg.openModal(syra_site, {
					id: helpers.uuid.generate(),
					content: this,
					$isAutoClose: false,
					$autoFocus: true,
					$center: true
				});
			}
		}
		// handling window close with ENTER or ESCAPE buttons
		this.keys = {};
		this.$item.$showDetail && this._bodyBtn.link.click();

		this.focusBtnId = this._defaultBtnId;
	},


	selectText: function() {
		if (this._coreFormatted) {
			var textItem = this._coreFormatted && syra_dom.isDisplay(this._coreFormatted) ? this._coreFormatted : this._coreRaw;
			if (document.body.createTextRange) { //ms
				var range = document.body.createTextRange();
				range.moveToElementText(textItem);
				range.select();
			} else
			if (window.getSelection) { //all others
				var selection = window.getSelection();
				var range = document.createRange();
				range.selectNodeContents(textItem);
				selection.removeAllRanges();
				selection.addRange(range);
			}
		}
	},
	drawPage: function() {
		var fHtmlEsc = helpers.string.htmlEscape;
		DesktopPage.prototype.drawPage.call(this);
		if (this.domTitle) {
			this.domTitle.className += " s-msgbox-msg-" + (this.$item.$severity || this.$item.$type);
		}
		this.domItem.syraItem = this.id;
		this.layoutSlot = this.domItem;
		this._message = this.boundFields.$message[0].domItem;
		var msg = ["<h1>", (fHtmlEsc(this.$item.$message || syra_local.msgbox_default_message)).replace(/\n/g, '</h1><h1>'), "</h1>"];
		$(this.boundFields.$message[0]._dataValue).html(msg.join(""));

		// links (AFTER MSG ADDITION !)
		if (this.$item.$links || this.$item.$actions) {
			this.$item.noDispDiag = true;
			this.applyChange(this.$item);
		}
		if (this.$item.$origin) {
			_ensureStackTrace(this);
			var origin = syra_local.msgbox_body_stackTrace_origin + this.$item.$origin;
			// add to formatted tab content
			$(syra_dom.addDiv("s-msgbox-origin", this._coreFormatted)).html(fHtmlEsc(origin));
			// add to raw tab content
			this._coreRaw.innerHTML += _preformat(this.$item.$origin + "<br/>");
		}
		if (this.$item.$details) {
			_ensureStackTrace(this);
			$(syra_dom.addDiv("s-msgbox-details", this._coreFormatted)).html((fHtmlEsc(this.$item.$details)).replace(/\n/g, '<br/>'));
			this._coreRaw.innerHTML += _preformat(this.$item.$details + "<br/>");
		}
		if (this.$item.$diagnoses && this.$item.$diagnoses.length > 0) {
			_addDiagnoses(this, this.$item.$diagnoses);
		}
		if (this.$item.$type == "prompt") {
			this.boundFields["$prompt"][0].input.setAttribute("size", 50);
		}
		this.getMenuItem(this._defaultBtnId).domItem.className += " s-msgbox-button-default";

		// handling window dimension
		this.domItem.style.minWidth = this.$item.$minWidth ? this.$item.$minWidth + "px" : "350px"; // "400px" ? "100px" ?
		this.$item.$isAutoClose && _addAutoClose(this);
	},
	onMenuClick: function(menuItem) {
		switch (menuItem.$bind) {
			case "yes":
			case "no":
			case "ok":
			case "cancel":
				if (this.$item.$useLinkAsOk && menuItem.$bind == "ok") {
					this.applyChange({
						$links: this.$item.$useLinkAsOk
					});
					var $bind = Object.keys(this.$item.$useLinkAsOk)[0];
					this.menuItems[$bind][0].domItem.click();
				} else {
					this.close(menuItem.$bind, {
						$clientId: menuItem.$bind,
						prompt: (this.$item.$type == "prompt") ? this.boundFields["$prompt"][0].getValue() : null
					});
				}
				return false;
		}
		return true;
	},


	resizeItem: function(setMinimumDimensions) {
		if (this._core) {
			this._core.style.height = "";
		}
		this.dialogWrapper && this.dialogWrapper.resizeDialog();

		if (this._core) {
			var diff = this.domItem.scrollHeight - this.domItem.clientHeight;
			if (diff) {
				this._core.style.height = Math.max((this._core.getBoundingClientRect().height - diff), 50) + "px";
			}
		}
	},
	applyEscape: function() {
		this.close("esc");
		return true;
	},
	close: function(closedBy, response) {
		if (this.$item) {
			this.$item.callback && this.$item.callback(response, closedBy);
			this.dispose();
		}
	},
	dispose: function() {
		if (this.$item) {
			delete this.$item.callback;
		}
		syra_dom.remove(this.domItem);
		if (this._groups) {
			var ids = Object.keys(this._groups);
			for (var ii = 0, jj = ids.length; ii < jj; ii++) {
				var id = ids[ii];
				this._groups[id].groupItem = null;
			}
		}
		if (this.autoCloseTimeOut) {
			clearInterval(this.autoCloseTimeOut);
		}
		this.dialogWrapper && this.dialogWrapper.dispose();
		DesktopPage.prototype.dispose.call(this);
	}
});
