// syracuse-ui/lib/alert/manager.js
"use strict";
var AlertDialog = require('syracuse-ui/lib/alert/alertDialog').AlertDialog;
var _alertPanel = require('syracuse-ui/lib/alert/alertPanel');

exports.page = function(error, $diagnoses) {
	switch (error.status) {
		case 401:
		case 403:
		case 404:
			syra_dom.hide(syra_site.body, true);
			syra_site.resetMainPage();
			var slot = syra_dom.addDiv("s-err-page");
			var container = syra_dom.addDiv("s-err-page-container", slot);
			var banner = syra_dom.addDiv("s-err-page-banner", container);
			var textSlot = syra_dom.addDiv("s-err-page-banner-text", banner);
			var node = textSlot.appendChild(document.createElement("pre"));
			node.className = "s-err-page-banner-title";
			node.textContent = syra_local["error_page_" + error.status];
			for (var ii = 0, jj = $diagnoses.length; ii < jj; ii++) {
				node = textSlot.appendChild(document.createElement("pre"));
				node.className = "s-err-page-banner-detail";
				node.textContent = $diagnoses[ii].$message;
			}
			syra_site.body.appendChild(slot);
			syra_dom.hide(syra_site.body, false);
			break;
		default:
			exports.dialog({
				$title: syra_local.error_page_default,
				$message: $diagnoses.length == 1 && $diagnoses[0].$message,
				$diagnoses: $diagnoses,
				callback: function() {
					if (syra_site.lastOpenedMainPageUrlSegments) {
						syra_url.history.load({
							$url: syra_site.lastOpenedMainPageUrlSegments.$url
						});
					} else {
						syra_menus.click.navigation();
					}
				}
			});
			return false;
	}
};
exports.box = {
	close: function(closedBy, id) {
		if (this.messageBox && (!id || id == this.messageBox.id)) {
			this.messageBox.close(closedBy);
		}
	},
	dispose: function() {
		this.messageBox && this.messageBox.dispose();
		this.messageBox = null;
	}
};

exports.dialog = function(options) {
	this.messageBox && this.messageBox.dispose();
	options.$type = options.$type || "alert";
	this.messageBox = new AlertDialog();
	this.messageBox.$prototype = {};
	syra_item.initialize(syra_site, this.messageBox, options);
	this.messageBox.loadBox();
	return this.messageBox;
};

exports.getDialog = function() {
	return this.messageBox;
};


exports.ask = function(options) {
	options.$type = "question";
	exports.dialog(options);
};


exports.clear = function(target) {
	exports.show(null, target);
};
/*{
 $bind: $bind
 autoHide
 $links
 noViewer
 }*/
exports.info = function(message, target, options) {
	exports.show(message, target, options, "info");
};
exports.warn = function(message, target, options) {
	exports.show(message, target, options, "warning");
};
exports.error = function(message, target, options) {
	exports.show(message, target, options, "error");
};
// null, string ,$message, [$mesage], $diagnoses
exports.show = function(message, target, options, severity) {
	if (message !== undefined) {
		if (message) {
			if (message.$diagnoses === undefined) {
				if (!Array.isArray(message)) {
					if (typeof message == "string") {
						message = {
							$message: message
						};
					}
					message = [message];
				}
				for (var ii = 0, jj = message.length; ii < jj; ii++) {
					message[ii].$severity = message[ii].$severity || severity || "error";
				}
				message = {
					$diagnoses: message
				};
			}
			message.$diagnoses = message.$diagnoses.length ? message.$diagnoses : null;
		} else {
			message = {
				$diagnoses: null
			};
		}
		if (target && target.isField) {
			message.field = target = target.variantItem ? target.variantItem : target;
			if (target.isDiagnoseDisabled) {
				return;
			}
			target.showDiagnosSlot && target.showDiagnosSlot();
			target = target.boxParent;
			if (message.$diagnoses && target.forceVisibility) {
				target.forceVisibility();
			}
		}
		var page = target && target.page;
		if (page && page.diagnosePage) {
			page = page.diagnosePage;
		}
		if (!page || page.disposed) {
			page = syra_dlg.getTopDialogPage();
			page = page && page._content;
			if (!page || page.disposed) {
				page = syra_site.mainPage;
			}
		}
		if (page && !page.disposed && page != syra_site && page != syra_site.userProfile) {
			if (!page.alertPanel) {
				page.alertPanel = _alertPanel.add(page);
			}
			page.alertPanel.showAlert(message, page, options);
		} else {
			if (message.$diagnoses) {
				var onlyOne = message.$diagnoses.length == 1 && message.$diagnoses[0];
				exports.box.show({
					$type: (onlyOne && onlyOne.$severity) || "alert",
					$buttons: "ok",
					$title: syra_local.siteMsgboxTitle,
					$diagnoses: onlyOne ? null : message.$diagnoses,
					$message: (onlyOne && onlyOne.$message) || syra_local.userProfile_siteMsgboxMsg,
					$origin: onlyOne ? onlyOne.$origin : "",
					$useLinkAsOk: (page && page == syra_site.userProfile) ? onlyOne && onlyOne.$links : null //temp hack
				});
			}
		}
	}
};
