// syracuse-ui/lib/alert/manager.js
"use strict";
var AlertModal = require('syracuse-ui/lib/alert/alertModal').AlertModal;
var _alertPanel = require('syracuse-ui/lib/alert/alertPanel');

var _modal;
exports.dispose = function() {
	_modal && _modal.dispose();
	_modal = null;
};

function _delayModal(options) {
	setTimeout(function() {
		delete options.$delay;
		exports.modal(options);
	}, options.$delay);
}

exports.modal = function(options) {
	if (options.$delay) {
		_delayModal(options);
		return;
	}
	_modal && _modal.dispose();
	options.$type = options.$type || "alert";
	_modal = new AlertModal();
	_modal.$prototype = {};
	syra_item.initialize(syra_site, _modal, options);
	_modal.load();
	return _modal;

};

exports.getModal = function() {
	return _modal;
};


exports.ask = function(options) {
	options.$type = options.$type || "question";
	return exports.modal(options);
};

exports.clear = function(target) {
	exports.show(null, target);
};

exports.info = function(message, target, options) {
	exports.show(message, target, options, "info");
};
exports.warn = function(message, target, options) {
	exports.show(message, target, options, "warning");
};
exports.error = function(message, target, options) {
	exports.show(message, target, options, "error");
};
// $bind: $bind in article applychange if field is not loaded
// null, string ,$message, [$mesage], $diagnoses
exports.show = function(message, target, options, severity) {
	if (message !== undefined) {
		if (message) {
			if (message.$diagnoses === undefined) {
				if (message.message && message.stack) {
					message = [{
						$message: message.message,
						$stackTrace: message.stack
					}];
				} else {
					if (!Array.isArray(message)) {
						if (typeof message == "string") {
							message = {
								$message: message
							};
						}
						message = [message];
					}
					for (var ii = 0, jj = message.length; ii < jj; ii++) {
						message[ii].$severity = message[ii].$severity || severity || "error";
					}
				}
				message = {
					$diagnoses: message
				};
			}
			message.$diagnoses = message.$diagnoses.length ? message.$diagnoses : null;
		} else {
			message = {
				$diagnoses: null
			};
		}
		if (target && target.isField) {
			message.field = target = target.variantItem ? target.variantItem : target;
			if (target.isDiagnoseDisabled) {
				return;
			}
			target.showDiagnosSlot && target.showDiagnosSlot();
			target = target.boxParent;
			if (message.$diagnoses && target.forceVisibility) {
				target.forceVisibility();
			}
		}
		var page = target && target.page;
		if (page && page.diagnosePage) {
			page = page.diagnosePage;
		}
		if (!page || page.disposed) {
			var over = syra_over.getMostOverPage();
			page = over && over.page;
			if (!page || page.disposed) {
				page = syra_site.mainPage;
			}
		}
		if (page && !page.disposed && page != syra_site && page != syra_site.userProfile) {
			if (!page.alertPanel) {
				page.alertPanel = _alertPanel.add(page);
			}
			if (options) {
				message.$links = options.$links; //controller.form
				message.autoHide = options.autoHide; //sap window
				message.autoHide = options.noViewer;
			}
			page.alertPanel.showAlert(message, page);
		} else {
			if (message.$diagnoses) {
				var onlyOne = message.$diagnoses.length == 1 && message.$diagnoses[0];
				exports.modal({
					$type: (onlyOne && onlyOne.$severity) || "alert",
					$buttons: "ok",
					$title: syra_local.siteMsgboxTitle,
					//$diagnoses: onlyOne ? null : message.$diagnoses,
					$diagnoses: message.$diagnoses,
					$stackTrace: onlyOne && onlyOne.$stackTrace,
					$message: (onlyOne && onlyOne.$message) || syra_local.userProfile_siteMsgboxMsg,
					$origin: onlyOne ? onlyOne.$origin : "",
					$useLinkAsOk: (page && page == syra_site.userProfile) ? onlyOne && onlyOne.$links : null //temp hack
				});
			}
		}
	}
};

exports.onUncaughtError = function(options) {
	var mainPage = syra_site.mainPage;
	if (mainPage && mainPage.onUncaughtError) {
		options.message = options.error.message;
		options.doEvent = function() {
			exports.error({
				$message: options.message, //can be overrided in page handler 
				$stackTrace: options.error.stack
			});
		};
		mainPage.onUncaughtError(options);
		return;
	}
	exports.error(options.error);
};
