// syracuse-ui/lib/mobile/mobileSiteGateway.js
"use strict";
var helpers = require("syracuse-core/lib/helpers");
var MobileJSONPage = require('syracuse-ui/lib/mobile/mobileJSONPage').MobileJSONPage;

function MobileSiteGateway() {}

exports.MobileSiteGateway = helpers.defineClass(MobileSiteGateway, null, {
	notify: function(event, eventData, info) {
		switch (event) {
			case "pagechange":
				if (!this.onChangeApplying) {
					this.resetData(eventData, info);
				}
				break;
			case "pageload":
				if (!this.onChangeApplying) {
					this.loadData(eventData, info);
				}
				this.onChangeApplying = false;
				break;
		}
		return eventData;
	},
	applyChange: function($article) {
		var cw = document.site.mobilePageFrame.contentWindow;
		if (this.name && cw.$.proto.auth.applyAuth) {
			this.onChangeApplying = true;
			cw.$.proto.auth.applyAuth(this.name, $article);
		}
	},
	resetData: function(eventData, info) {
		this.name = "";
	},
	loadData: function(data, info) {
		this.name = info.name;
		this.$authorUrl = info.$authorUrl;
		var httpQuery = document.controller.parseUrl(info.url.origUri);
		var $itemPage = {
			httpQuery: httpQuery,
			$urlParts: httpQuery.$urlParts,
			$representation: {
				$authorUrl: info.$authorUrl,
				$prototype: info.$prototype,
				$article: data
			}
		};
		document.site.onMainPageChange($itemPage);
	},
	onShowTargetPageDiagnoses: function(message, options) {
		if (this.jsonPage) {
			this.jsonPage.showDiagnoses(message, options);
		}
	},
	toggleJsonEditor: function(show) {
		var mainPage = document.site.mainPage;
		if (show) {
			mainPage._item.style.display = "none";
			this.jsonPage = new MobileJSONPage();
			this.jsonPage.localize = document.site.localize;
			this.jsonPage._initializePage();
			this.jsonPage.layoutSlot = document.createElement("div");
			document.site.body.appendChild(this.jsonPage.layoutSlot);
			this.jsonPage.loadBox();
			this.jsonPage.fillEditors(mainPage.$prototype, mainPage.$item);
		} else {
			mainPage._item.style.display = "";
			if (this.jsonPage) {
				if (this.jsonPage.layoutSlot) {
					document.site.removeDomChild(this.jsonPage.layoutSlot);
				}
				document.controller.disposeObject(this.jsonPage);
				this.jsonPage = null;
			}
		}
	}
});
