// syracuse-ui/lib/mobile/mobileSite.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Site = require('syracuse-ui/lib/site/site').Site;
var MobilePage = require('syracuse-ui/lib/mobile/mobilePage').MobilePage;

function MobileSiteGateway(){
}

exports.MobileSiteGateway = helpers.defineClass(MobileSiteGateway, null, {
    notify: function(event, eventData, info){
        switch (event) {
            case "pagechange":
                this.resetData(eventData, info);
                break;
            case "pageload":
                this.loadData(eventData, info);
                break;
        }
        return eventData;
    },
    resetData: function(eventData, info){
        this.name = "";
    },
    loadData: function(data, info){
        this.name = info.name;
        this.$authorUrl = info.$authorUrl;
        var httpQuery = document.controller.parseUrl(info.url.origUri);
        var $itemPage = {
            httpQuery: httpQuery,
            $urlParts: httpQuery.$urlParts,
            $representation: {
                $authorUrl: info.$authorUrl,
                $prototype: info.$prototype
            }
        };
        document.site.onMainPageChange($itemPage);
    }
});
function MobileSite(){
}

exports.MobileSite = helpers.defineClass(MobileSite, Site, {
    dispose: function(){
        this.mobileGateway = null;
        Site.prototype.dispose.call(this);
    },
    _releaseMainPage: function(){
        Site.prototype._releaseMainPage.call(this);
    },
    
    onMainPageChange: function($itemPage){
        if (!this.body) {
            this.body = this.$$body[0]; //hack temp for office
        }
        this.body.style.display = "none";
        $itemPage.boxParent = null;
        $itemPage.layoutSlot = this.$$body[0];
        this._releaseMainPage();
        if (this._diagnosesPanel) {
            this._diagnosesPanel.emptyViewer();
        }
        this.toggleTopPanel(false);
        
        this.mainPage = this.loadNewPage($itemPage);
        this.mainPage.openerHttpQuery = $itemPage.httpQuery;
        this.mainPage.isMainPage = true;
        this.updateDocumentTitle();
        this.body.style.display = "";
        this.resize();
        this.spyGateway.setSpyedPage(this.mainPage);
        return this.mainPage;
    },
    drawBox: function(){
        var self = this;
        $.auth = self.mobileGateway = new MobileSiteGateway();
        self._renderHeader();
        self.$$item = self.$$body = $(self.body = document.createElement("div"));
        self._$$header = $(self._header);
        self._mobileWrapper = document.createElement("div");
        self._mobileWrapper.setAttribute("id", "s-mobile-site-wrapper");
        self.mobilePageSlot = document.createElement("div");
        self.mobilePageSlot.setAttribute("id", "s-mobile-page-slot");
        self.body.setAttribute("id", "s-site-body");
        self.body.setAttribute("tabindex", "1"); // add for receive keyboard event  for shortcut
        self._mobileWrapper.appendChild(self.mobilePageSlot);
        self._mobileWrapper.appendChild(self.body);
        self.layoutSlot.appendChild(self._mobileWrapper);
        
        self.mobilePageFrame = document.createElement("iframe");
        self.mobilePageFrame.className = "s-mobile-page-frame";
        self.mobilePageFrame.onload = function(dd, dss, dsd){
            var cw = self.mobilePageFrame.contentWindow;
            if (cw.$.proto.auth.applyAuth) {
                //  cw.$.proto.auth.applyAuth($auth.name, value);
            }
        };
        
        self.mobilePageSlot.appendChild(self.mobilePageFrame);
        
        self.logon(function(){
            document.controller.startNavigation();
            self.mobilePageFrame.setAttribute("src", "mobile-debug.html");
            self._onLogon();
            self.layoutSlot.style.display = "";
            self.resize();
        });
    },
    _renderHeader: function(){
        this._header = document.createElement("header");
        this._header.setAttribute("id", "s-site-header");
        this.setZIndex(this._header);
        this.layoutSlot.appendChild(this._header);
        this._topPanelContainer = document.createElement("div");
        this._topPanelContainer.setAttribute("id", "s-site-top-pn");
        this.layoutSlot.appendChild(this._topPanelContainer);
        var topLeft = document.createElement("div");
        topLeft.setAttribute("id", "s-site-header-top-left");
        this._header.appendChild(topLeft);
        
        var div = document.createElement("div");
        div.setAttribute("id", "s-site-sage");
        topLeft.appendChild(div);
        
        div = document.createElement("div");
        div.setAttribute("id", "s-site-help");
        this.loadNewItem(topLeft.appendChild(div), {
            $id: "s-site-help-link",
            $bind: "$help",
            $category: "link",
            $noText: true,
            $skin: "s-site-help-link"
        });
        this._appendTopPanelOpener(topLeft);
        
        this._authoringLink = document.createElement("a");
        this._authoringLink.className = "s-site-authoring-link";
        this._authoringLink.setAttribute("href", "#");
        this._authoringLink.setAttribute("data-s-picker", "s-site-authoring-link");
        topLeft.appendChild(this._authoringLink);
        
        this.logoutSlot = document.createElement("div");
        this.logoutSlot.setAttribute("id", "s-site-logout");
        topLeft.appendChild(this.logoutSlot);
        
    }
});

exports.load = function($item, $prototype){
    var widgetsLibrary = require('syracuse-ui/lib/site/widgetsLibrary');
    widgetsLibrary.defaultPageCategory = MobilePage;
    return (new MobileSite()).loadBox({
        widgetsLibrary: widgetsLibrary,
        userProfileClass: require('syracuse-ui/lib/site/userProfile').UserProfile,
        fusionGatewayClass: require('syracuse-ui/lib/fusion/fusionGateway').FusionGateway,
        $item: $item,
        $prototype: $prototype
    });
};
