// syracuse-ui/lib/mobile/mobileSiteGateway.js
"use strict";
var helpers = require("syracuse-core/lib/helpers");
var MobileJSONPage = require('syracuse-ui/lib/mobile/mobileJSONPage').MobileJSONPage;

function MobileSiteGateway() {}

exports.MobileSiteGateway = helpers.defineClass(MobileSiteGateway, null, {
	notify: function(event, eventData, info) {
		switch (event) {
			case "pagechange":
				if (!this.onChangeApplying) {
					this.resetData(eventData, info);
				}
				break;
			case "pageload":
				if (!this.onChangeApplying) {
					this.loadData(eventData, info);
				}
				this.onChangeApplying = false;
				break;
		}
		return eventData;
	},
	applyChange: function($article) {
		if (!this.isApplyChangeDisabled) {
			var cw = syra_site.mobilePageFrame.contentWindow;
			if (this.name && cw.$.proto.auth.applyAuth) {
				this.onChangeApplying = true;
				cw.$.proto.auth.applyAuth(this.name, $article);
			}
		}
	},
	resetData: function(eventData, info) {
		this.name = "";
	},
	loadData: function(data, info) {
		this.name = info.name;
		syra_site.onMainPageChange({
			openerUrlSegments: syra_site.urlMaker.parse(info.url.origUri),
			initData: info.dataset,
			$representation: {
				$authorUrl: this.$authorUrl = info.$authorUrl,
				$prototype: info.$prototype,
				$article: data
			}
		});
	},
	onShowTargetPageDiagnoses: function(message, options) {
		if (this.jsonPage) {
			this.jsonPage.showDiagnoses(message, options);
		}
	},
	toggleJsonEditor: function(show) {
		var mainPage = syra_site.mainPage;
		if (show) {
			mainPage.domItem.style.display = "none";
			this.jsonPage = new MobileJSONPage();
			syra_site.pageLoader.initialize(this.jsonPage);
			this.jsonPage.layoutSlot = document.createElement("div");
			syra_site.body.appendChild(this.jsonPage.layoutSlot);
			this.jsonPage.loadBox();
			this.jsonPage.fillEditors(mainPage.$prototype, mainPage.$item);
			this.jsonPage.isPageLoaded = true;
			this.jsonPage.ensurePageVisibility();
		} else {
			mainPage.domItem.style.display = "";
			if (this.jsonPage) {
				if (this.jsonPage.layoutSlot) {
					syra_site.dom.removeChild(this.jsonPage.layoutSlot);
				}
				this.jsonPage.dispose();
				this.jsonPage = null;
			}
		}
	}
});
