// syracuse-ui/lib/mobile/mobileSite.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Site = require('syracuse-ui/lib/site/site').Site;
var MobilePage = require('syracuse-ui/lib/mobile/mobilePage').MobilePage;
var MobileSiteGateway = require('syracuse-ui/lib/mobile/mobileSiteGateway').MobileSiteGateway;

function MobileSite() {}

exports.MobileSite = helpers.defineClass(MobileSite, Site, {
	dispose: function() {
		this.mobileGateway = null;
		Site.prototype.dispose.call(this);
	},
	_releaseMainPage: function() {
		Site.prototype._releaseMainPage.call(this);
	},
	onMainPageChange: function($itemPage) {
		this.body.style.display = "none";
		$itemPage.boxParent = null;
		$itemPage.layoutSlot = this.body;
		this._releaseMainPage();
		if (this.diagnosesPanel) {
			this.diagnosesPanel.clean();
		}
		this.toggleTopPanel(false);

		this.mainPage = this.loadNewPage($itemPage);
		this.mainPage.openerUrlSegments = $itemPage.openerUrlSegments;
		this.mainPage.isMainPage = true;
		this.updateDocumentTitle();
		this.pageDesignerOpener.style.visibility = "";
		this.body.style.display = "";
		this.resize();
		this.spyGateway.setSpyedPage(this.mainPage);
		return this.mainPage;
	},
	drawBox: function() {
		var self = this;
		self.isMobileSite = true;
		self.addDragDropManager();
		$.auth = self.mobileGateway = new MobileSiteGateway();
		self._renderHeader();
		self.$$item = $(self.body = document.createElement("div"));
		self._mobileWrapper = document.createElement("div");
		self._mobileWrapper.setAttribute("id", "s-mobile-site-wrapper");
		self.mobilePageSlot = document.createElement("div");
		self.mobilePageSlot.setAttribute("id", "s-mobile-page-slot");
		self.body.setAttribute("id", "s-site-body");
		self.body.setAttribute("tabindex", "1"); // add for receive keyboard event  for shortcut
		self._mobileWrapper.appendChild(self.mobilePageSlot);
		self._mobileWrapper.appendChild(self.body);
		self.layoutSlot.appendChild(self._mobileWrapper);

		self.mobilePageFrame = document.createElement("iframe");
		self.mobilePageFrame.className = "s-mobile-page-frame";
		self.mobilePageSlot.appendChild(self.mobilePageFrame);

		self.logon(function() {
			self.mobilePageFrame.setAttribute("src", "/syracuse-mobile/html/mobile-debug.html");
			self._onLogon();
			self.layoutSlot.style.display = "";
			self.resize();
		});
	},
	_renderHeader: function() {
		this.header = document.createElement("header");
		this.header.id = "s-site-header";
		this.dom.setZIndex(this.header);
		this.layoutSlot.appendChild(this.header);

		this._topPanelContainer = document.createElement("div");
		this._topPanelContainer.setAttribute("id", "s-site-top-pn");
		this.layoutSlot.appendChild(this._topPanelContainer);

		this.headerTop = document.createElement("div");
		this.headerTop.id = "s-site-header-top";

		this.headerLogo = document.createElement("a");
		this.headerLogo.id = "s-site-header-logo";
		this.headerLogo.setAttribute("data-s-picker", "s-site-sage");
		this.headerTop.appendChild(this.headerLogo);

		this.pageDesignerOpener = document.createElement("a");
		this.pageDesignerOpener.style.visibility = "hidden";
		this.pageDesignerOpener.className = "s-page-designer-opener";
		this.pageDesignerOpener.setAttribute("href", "#");
		this.pageDesignerOpener.setAttribute("data-s-picker", "s-page-designer-opener");
		this.headerTop.appendChild(this.pageDesignerOpener);

		var middleCell = document.createElement("div");
		middleCell.id = "s-site-header-middle-cell";
		this.headerTop.appendChild(middleCell);


		this._appendTopPanelOpener(middleCell);

		this.appendLogoutButton();

		this.loadNewItem(this.headerTop, {
			$clientId: "s-site-help-link",
			$bind: "$help",
			$category: "link",
			$noText: true,
			$skin: "s-site-help-link"
		});

		this.header.appendChild(this.headerTop);

	}
});

exports.load = function($item, $prototype) {
	var widgetsLibrary = require('syracuse-ui/lib/site/widgetsLibrary');
	widgetsLibrary.defaultPageCategory = MobilePage;
	return (new MobileSite()).loadBox({
		widgetsLibrary: widgetsLibrary,
		fusionGatewayClass: require('syracuse-ui/lib/fusion/fusionGateway').FusionGateway,
		$item: $item,
		$prototype: $prototype
	});
};
