// syracuse-ui/lib/mobile/mobileJSONPage.js
"use strict";
var helpers = require("syracuse-core/lib/helpers");
var DesktopPage = require("syracuse-ui/lib/page/desktopPage").DesktopPage;
var locale = require('syracuse-core/lib/locale');

function MobileJSONPage() {}

exports.MobileJSONPage = helpers.defineClass(MobileJSONPage, DesktopPage, {
	loadBox: function($fieldsPrototype, initData, $initDiagnoses) {
		this.isAuthoringEventEnabled = true;
		this.$pageCategory = "page";
		this.mobileLocalize = locale.resources(module)();
		this.$prototype = {
			$properties: {
				$fieldPrototype: {
					"$type": "application/x-string"
				},
				$fieldJSON: {
					"$type": "application/x-string"
				}
			},
			$actions: {
				$applyChange: {
					$title: this.mobileLocalize.p_applychange
				}
			}
		};
		this.$item = {
			$layout: {
				$items: [{
					$skin: "s-record-menus-link",
					$category: "link",
					$bind: "$applyChange"

				}, {
					$layoutType: "tabs",
					$items: [{
						$category: "section",
						$title: this.mobileLocalize.p_fieldJSON,
						$layout: {
							$items: [{
								$bind: "$fieldJSON",
								$isTitleHidden: true,
								$isEditMode: true,
								$rows: 50,
								$isAutoSizeDisabled: true,
								$skin: "s-json-field"
							}]
						}

					}, {
						$category: "section",
						$title: this.mobileLocalize.p_fieldPrototype,
						$layout: {
							$items: [{
								$bind: "$fieldPrototype",
								$isTitleHidden: true,
								$isEditMode: true,
								$rows: 50,
								$isAutoSizeDisabled: true,
								$skin: "s-json-field"
							}]
						}
					}]
				}]
			}
		};
		this.$autoFetch = false;
		DesktopPage.prototype.loadBox.call(this, initData, $initDiagnoses);
		this._header.style.display = "none";
		this._body.style.padding = "0.5em";
	},
	notifyActionChange: function(target, value) {
		if (value.$applyChange) {
			var diag = null;
			try {
				var authorPage = document.site.authorPage;
				var field = this.boundFields.$fieldJSON[0];
				var $article = JSON.parse(field.getDataValue());
				field.setInputValue(JSON.stringify($article, null, 2));
				authorPage.designedPage.$item = $article;
				document.site.mobileGateway.isApplyChangeDisabled = true;
				authorPage.designedPage.designer.tools.history.notifyUpdate(true);
				delete document.site.mobileGateway.isApplyChangeDisabled;
				authorPage.designedPage.designer.onEndChangeStep(helpers.object.clone($article, true));
				diag = {
					$message: this.mobileLocalize.p_changeapplied,
					$severity: "success"
				};
			} catch (error) {
				diag = {
					$message: error.message
				};
			} finally {
				delete document.site.mobileGateway.isApplyChangeDisabled;
				this.showDiagnoses({
					autoHide: diag.$severity == "success" ? {
						autoHideTimeOut: 3000
					} : null,
					$diagnoses: [diag]
				});
			}
			return;
		}
		DesktopPage.prototype.notifyActionChange.call(this, target, value);
	},
	fillEditors: function($prototype, $article) {
		this.applyChange({
			$fieldJSON: JSON.stringify($article, null, 2),
			$fieldPrototype: JSON.stringify($prototype, null, 2)

		});
	}
});
