// syracuse-ui/lib/mobile/mobileSite.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Site = require('syracuse-ui/lib/site/site').Site;
var MobilePage = require('syracuse-ui/lib/mobile/mobilePage').MobilePage;
var MobileSiteGateway = require('syracuse-ui/lib/mobile/mobileSiteGateway').MobileSiteGateway;

function MobileSite() {}

exports.MobileSite = helpers.defineClass(MobileSite, Site, {
	dispose: function() {
		this.mobileGateway = null;
		Site.prototype.dispose.call(this);
	},
	_releaseMainPage: function() {
		Site.prototype._releaseMainPage.call(this);
	},

	onMainPageChange: function($itemPage) {
		if (!this.body) {
			this.body = this.$$body[0]; //hack temp for office
		}
		this.body.style.display = "none";
		$itemPage.boxParent = null;
		$itemPage.layoutSlot = this.$$body[0];
		this._releaseMainPage();
		if (this.diagnosesPanel) {
			this.diagnosesPanel.clean();
		}
		this.toggleTopPanel(false);

		this.mainPage = this.loadNewPage($itemPage);
		this.mainPage.openerHttpQuery = $itemPage.httpQuery;
		this.mainPage.isMainPage = true;
		this.updateDocumentTitle();
		this.pageDesignerOpener.style.visibility = "";
		this.body.style.display = "";
		this.resize();
		this.spyGateway.setSpyedPage(this.mainPage);
		return this.mainPage;
	},
	drawBox: function() {
		var self = this;
		self.isMobileSite = true;
		$.auth = self.mobileGateway = new MobileSiteGateway();
		self._renderHeader();
		self.$$item = self.$$body = $(self.body = document.createElement("div"));
		self._$$header = $(self.header);
		self._mobileWrapper = document.createElement("div");
		self._mobileWrapper.setAttribute("id", "s-mobile-site-wrapper");
		self.mobilePageSlot = document.createElement("div");
		self.mobilePageSlot.setAttribute("id", "s-mobile-page-slot");
		self.body.setAttribute("id", "s-site-body");
		self.body.setAttribute("tabindex", "1"); // add for receive keyboard event  for shortcut
		self._mobileWrapper.appendChild(self.mobilePageSlot);
		self._mobileWrapper.appendChild(self.body);
		self.layoutSlot.appendChild(self._mobileWrapper);

		self.mobilePageFrame = document.createElement("iframe");
		self.mobilePageFrame.className = "s-mobile-page-frame";
		self.mobilePageSlot.appendChild(self.mobilePageFrame);

		self.logon(function() {
			self.mobilePageFrame.setAttribute("src", "/syracuse-mobile/html/mobile-debug.html");
			self._onLogon();
			self.layoutSlot.style.display = "";
			self.resize();
		});
	},
	_renderHeader: function() {
		this.header = document.createElement("header");
		this.header.setAttribute("id", "s-site-header");
		this.setZIndex(this.header);
		this.layoutSlot.appendChild(this.header);
		this._topPanelContainer = document.createElement("div");
		this._topPanelContainer.setAttribute("id", "s-site-top-pn");
		this.layoutSlot.appendChild(this._topPanelContainer);
		var topLeft = document.createElement("div");
		topLeft.setAttribute("id", "s-site-header-top-left");
		this.header.appendChild(topLeft);

		var div = document.createElement("div");
		div.setAttribute("id", "s-site-sage");
		topLeft.appendChild(div);

		div = document.createElement("div");
		div.setAttribute("id", "s-site-help");
		this.loadNewItem(topLeft.appendChild(div), {
			$id: "s-site-help-link",
			$bind: "$help",
			$category: "link",
			$noText: true,
			$skin: "s-site-help-link"
		});
		this._appendTopPanelOpener(topLeft);

		this.pageDesignerOpener = document.createElement("a");
		this.pageDesignerOpener.style.visibility = "hidden";
		this.pageDesignerOpener.className = "s-page-designer-opener";
		this.pageDesignerOpener.setAttribute("href", "#");
		this.pageDesignerOpener.setAttribute("data-s-picker", "s-page-designer-opener");
		topLeft.appendChild(this.pageDesignerOpener);

		var logoutSlot = document.createElement("div");
		logoutSlot.setAttribute("id", "s-site-logout");
		this.loadNewItem(topLeft.appendChild(logoutSlot), {
			$bind: "$logout",
			$category: "link",
			$skin: "s-user-profile-logout-top"
		});
	}
});

exports.load = function($item, $prototype) {
	var widgetsLibrary = require('syracuse-ui/lib/site/widgetsLibrary');
	widgetsLibrary.defaultPageCategory = MobilePage;
	return (new MobileSite()).loadBox({
		widgetsLibrary: widgetsLibrary,
		userProfileClass: require('syracuse-ui/lib/site/userProfile').UserProfile,
		fusionGatewayClass: require('syracuse-ui/lib/fusion/fusionGateway').FusionGateway,
		$item: $item,
		$prototype: $prototype
	});
};
