// syracuse-ui/lib/authoring/page/designProperties.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function _getLayoutSection($level) {
	var $layout = {
		$category: "section",
		$skin: "s-aw-h2-" + $level,
		$isSectionModels: true,
		$isStructure: true,
		$isEmptyVisible: true,
		$title: syra_local.aw_template,
		$layout: {
			$items: []
		}
	};
	return $layout;
}

function _getAppearanceSection($level) {
	return {
		$category: "section",
		$skin: "s-aw-h2-" + $level,
		$isTabChoice: "modelTab",
		$isStructure: true,
		$isEmptyVisible: true,
		$title: syra_local.aw_appearance,
		$layout: {
			$items: []
		}
	};
}

function _getRowSection() {
	return {
		$category: "section",
		$skin: "s-aw-h2-layout",
		$isRowSection: true,
		$isSectionModels: true,
		$isColumnSection: true,
		$isEmptyVisible: true,
		$title: syra_local.aw_rowLayout,
		$layout: {
			$items: [{
				$category: "field",
				$bind: "$rowAlign",
				$css: "s-aw-prop",
				$isEditMode: true
			}]
		}
	};
}



exports.fillRowSection = function(section) {
	section.dataStore = section.dataStore || {};
	if (!section.dataStore.insertRow) {
		var row = document.createElement("div");
		row.className = "s-aw-insert-row";

		var cell = document.createElement("div");
		cell.className = "s-aw-insert-row-label";
		cell.textContent = syra_local.aw_addrow;
		row.appendChild(cell);

		cell = document.createElement("div");
		cell.className = "s-aw-insert-row-links";
		var links = ["addRowBefore", "addRowAfter"];
		for (var ii = 0, jj = links.length; ii < jj; ii++) {
			var link = document.createElement("a");
			link.textContent = syra_local["aw_" + links[ii]];
			link.className = "s-aw-insert-row-link";
			link.syraOnClick = "onAddRowClick";
			link.syraAdd = links[ii];
			cell.appendChild(link);
		}
		row.appendChild(cell);
		section.body.insertBefore(section.dataStore.insertRow = row, section.body.firstChild);
	}
};

exports.article = function() {
	return {
		$category: "section",
		$skin: "s-aw-h1-article",
		$title: syra_local.aw_page,
		$isHidden: true,
		$layout: {
			$layoutType: "stack",
			$items: [_getRowSection(), _getLayoutSection("article"), {
				$category: "section",
				$skin: "s-aw-h2-article",
				$bind: "$field",
				$isSectionFields: true,
				$title: syra_local.aw_sectionFields,
				$layout: {
					$layoutType: "stack",
					$items: [{
						$bind: "$fieldsIsTitleHidden",
						$css: "s-aw-prop",
						$isEditMode: true,
						$isLeftSpaceHidden: true
					}, {
						$bind: "$labelAlignment",
						$css: "s-aw-prop",
						$isEditMode: true,
						$choiceLayout: "1"
					}]
				}
			}]
		}
	};
};


exports.section = function() {
	return {
		$category: "section",
		$skin: "s-aw-h1-section",
		$title: syra_local.aw_section,
		$isHidden: true,
		$layout: {
			$layoutType: "stack",
			$items: [_getRowSection(), _getAppearanceSection("section"), _getLayoutSection("section"), {
				$category: "section",
				$skin: "s-aw-h2-section",
				$isTitleSection: true,
				$title: syra_local.aw_sectionAppearance,
				$layout: {
					$items: [{
						$bind: "$title",
						$css: "s-aw-prop",
						$isEditMode: true,
						$isLeftSpaceHidden: true
					}, {
						$bind: "$titleChoice",
						$css: "s-aw-prop",
						$isEditMode: true,
						$choiceLayout: "1"
					}, {
						$bind: "$titleLocalization",
						$css: "s-aw-prop",
						$isEditMode: true,
						$format: "$combo"
					}]
				}
			}, {
				$category: "section",
				$skin: "s-aw-h2-section",
				$title: syra_local.aw_sectionBehaviour,
				$layout: {
					$items: [{
						$bind: "$isBoxCollapsable",
						$css: "s-aw-prop",
						$isEditMode: true,
						$isLeftSpaceHidden: true
					}, {
						$bind: "$isMaximizable",
						$css: "s-aw-prop",
						$isEditMode: true,
						$isLeftSpaceHidden: true
					}]
				}
			}, {
				$category: "section",
				$skin: "s-aw-h2-section",
				$bind: "$field",
				$isSectionFields: true,
				$title: syra_local.aw_sectionFields,
				$layout: {
					$layoutType: "stack",
					$items: [{
						$bind: "$fieldsIsTitleHidden",
						$css: "s-aw-prop",
						$isEditMode: true,
						$isLeftSpaceHidden: true
					}, {
						$bind: "$labelAlignment",
						$css: "s-aw-prop",
						$isEditMode: true,
						$choiceLayout: "1"
					}]
				}
			}]
		}
	};
};

exports.block = function() {
	return {
		$category: "section",
		$skin: "s-aw-h1-block",
		$title: syra_local.aw_block,
		$isHidden: true,
		$layout: {
			$layoutType: "stack",
			$items: [_getRowSection(), _getAppearanceSection("block"), _getLayoutSection("block"), {
				$category: "section",
				$skin: "s-aw-h2-block",
				$isTitleSection: true,
				$title: syra_local.aw_sectionAppearance,
				$layout: {
					$items: [{
						$bind: "$title",
						$css: "s-aw-prop",
						$isEditMode: true,
						$isLeftSpaceHidden: true
					}, {
						$bind: "$titleChoice",
						$css: "s-aw-prop",
						$isEditMode: true,
						$choiceLayout: "1"
					}, {
						$bind: "$titleLocalization",
						$css: "s-aw-prop",
						$isEditMode: true,
						$format: "$combo"
					}]
				}
			}, {
				$category: "section",
				$skin: "s-aw-h2-block",
				$title: syra_local.aw_sectionBehaviour,
				$layout: {
					$items: [{
						$bind: "$isBoxCollapsable",
						$css: "s-aw-prop",
						$isEditMode: true,
						$isLeftSpaceHidden: true
					}, {
						$bind: "$isMaximizable",
						$css: "s-aw-prop",
						$isEditMode: true,
						$isLeftSpaceHidden: true
					}]
				}
			}, {
				$category: "section",
				$skin: "s-aw-h2-block",
				$bind: "$field",
				$isSectionFields: true,
				$title: syra_local.aw_sectionFields,
				$layout: {
					$layoutType: "stack",
					$items: [{
						$bind: "$fieldsIsTitleHidden",
						$css: "s-aw-prop",
						$isEditMode: true,
						$isLeftSpaceHidden: true
					}, {
						$bind: "$labelAlignment",
						$css: "s-aw-prop",
						$isEditMode: true,
						$choiceLayout: "1"
					}]
				}
			}]
		}
	};
};

exports.field = function() {
	return {
		$category: "section",
		$skin: "s-aw-h1-field",
		$title: syra_local.aw_field,
		$isHidden: true,
		$layout: {
			$layoutType: "stack",
			$items: [{
					$bind: "$isAdvanced",
					$css: "s-aw-prop",
					$isEditMode: true,
					$isLeftSpaceHidden: true
				},
				_getRowSection(), _getLayoutSection("field"), {
					$category: "section",
					$skin: "s-aw-h2-field",
					$isLabelSection: true,
					$title: syra_local.aw_sectionFieldLabel,
					$layout: {
						$layoutType: "stack",
						$items: [{
							$bind: "$isTitleHidden",
							$title: syra_local.aw_isLabelHidden,
							$css: "s-aw-prop",
							$isEditMode: true,
							$isLeftSpaceHidden: true
						}, {
							$bind: "$labelAlignment",
							$title: syra_local.aw_sectionFieldLabelPosition,
							$css: "s-aw-prop",
							$isEditMode: true,
							$choiceLayout: "1"
						}]
					}
				}, {
					$category: "section",
					$title: syra_local.aw_sectionFieldWidget,
					$skin: "s-aw-h2-field",
					$isHidden: true,
					$isFieldAddinSection: true,
					$layout: {
						$items: []
					}
				}
			]
		}
	};
};
