// syracuse-ui/lib/search/searchPage.js
"use strict";
var helpers = require("syracuse-core/lib/helpers");
var DesktopPage = require('syracuse-ui/lib/page/desktopPage').DesktopPage;
var _facetsBar = require('./facetsBar');

var _$searchLinks = {
	$startWith: {
		$url: "/sdata/syracuse/search/syracuse/$search?representation=queryResult.$search&q={$search}"
	},
	$similar: {
		$url: "/sdata/syracuse/search/syracuse/$search?representation=queryResult.$search&q={$search}&qt=fuzzy"
	}
};

function _sendOneSearchRequest(page, dataset, options, isDataSource) {
	page.applyChange({
		query: page.openerUrlSegments.params.q
	});
	syra_site.urlMaker.formatMenuUrl(page, {
		$url: _$searchLinks[page._modeBtn.syraMode].$url + (isDataSource ? "&source=data" : "&source=functions")
	}, dataset, function($url, isCanceled) {
		options.$location = options.$location || {
			$url: $url
		};
		if (!page.isMultiQuery) {
			options.$updateHitory = true;
		}

		syra_controller.callServer(page, options, function(data) {
			var queryCode = isDataSource ? "$queryData" : "$queryFunctions";
			page.applyChange({
				query: data.query,
				$searchFacets: isDataSource ? data.$searchFacets : page.dataset.$searchFacets // no search facet for 'functions' results
			});
			// apply vignette pages data (the tabs)
			var $resources = data[queryCode] = data.$resources;
			delete data.$resources;

			if (data.$totalResults != undefined) {
				var key = isDataSource ? "searchDataTab" : "searchFunctionsTab";
				page.idMap[key].setTitle(syra_local[key] + " (" + data.$totalResults + ")");
			}
			for (var ii = 0, jj = $resources.length; ii < jj; ii++) {
				$resources[ii].$descriptionData = [];
				if ($resources[ii].$properties) {
					_appendDescription($resources[ii].$properties, $resources[ii], $resources[ii], queryCode);
				}
			}
			page.boundFields[isDataSource ? "$vignetteData" : "$vignetteFunctions"][0].vignette.applyChange(data);
			_facetsBar.checkVisibility();
		});
	});
}

function _sendAllSearchRequests(page, dataset, options) {
	page.isMultiQuery = true;
	// to handle navigation (prev/next) issues.
	var functionsSourceOptions = helpers.object.clone(options);
	var dataSourceOptions = helpers.object.clone(options);
	// remove params unecessary for appropriate source
	if (page.openerUrlSegments.params.source) {
		if (page.openerUrlSegments.params.source == 'data') {
			delete functionsSourceOptions.params;
		} else {
			delete dataSourceOptions.params;
		}
	}
	_sendOneSearchRequest(page, dataset, functionsSourceOptions, false, page.openerUrlSegments.params.source);
	_sendOneSearchRequest(page, dataset, dataSourceOptions, true, page.openerUrlSegments.params.source);
	page.isMultiQuery = false;
}

function _appendDescription($properties, recordData, parentRecordData, queryCode) {
	var $binds = Object.keys($properties);
	for (var mm = 0, pp = $binds.length; mm < pp; mm++) {
		var property = $properties[$binds[mm]];
		if ((property.$isHighlight || property.$isSummary) && recordData[$binds[mm]]) {
			var span = document.createElement("span");
			span.className = 's-search-res-desc-title';
			span.textContent = property.$title + ": ";
			var value = document.createElement("span");
			value.className = 's-search-res-desc-data';
			var data = recordData[$binds[mm]];
			if (data) {
				if (!data.replace && data.join) {
					data = data.join();
				}
				if (data.replace) {
					value.innerHTML = syra_site.dom.formatHTMLText(data.replace(/<em>/g, "<em class='s-search-res-data-em'>"), false);
				}
			}
			parentRecordData.$descriptionData.push(span.outerHTML + value.outerHTML);
		} else {
			if (property.$isHighlight == undefined) {
				_appendDescription(property, recordData[$binds[mm]], parentRecordData);
			}
		}
	}
}

function _getVignetteProto($bind) {
	var $prototype = {
		$article: {
			$layout: {
				$layoutType: "stack",
				$items: [{
					$bind: $bind,
					$selectMode: "row",
					$executeUrl: true,
					$isMenuRecordHidden: true,
					$alternateStyle: false,
					$noDataText: syra_local.search_no_res,
					$format: "cards",
					$skin: "s-search-res",
					$css: "s-search-res",
					$addMaximize: true,
					renderRecordContent: function(record) {
						_addRecordText(record, "$searchItemTitle", "s-search-item-title");
						$bind == "$queryData" && _addRecordText(record, "$resultTypeTitle", "s-search-item-type");
						var list = record.dataset.$descriptionData;
						for (var ii = 0, jj = list.length; ii < jj; ii++) {
							var div = document.createElement("div");
							div.className = "s-search-item-desc";
							div.innerHTML = list[ii];
							record.body.appendChild(div);
						}
					},
					$layout: {
						$items: []
					}
				}]
			}
		},
		$properties: {}
	};
	$prototype.$properties[$bind] = {
		$type: "application/x-array",
		$item: {
			$type: "application/json",
			$prototype: "{$baseUrl}/$prototypes('{$representation}.$search')",
			$url: "{$baseUrl}/{$pluralType}('{$key}')",
			$value: "{$key}",
			$title: "queryResult {$key}",
			$key: "{$uuid}",
			$description: "queryResult {$key}",
			$pluralType: "queryResults",
			$representation: "queryResult",
			$defaultOrder: "title",
			$properties: {
				$searchItemTitle: {
					$type: "application/x-string",
					$format: "$html",
					$links: {
						"$details": {
							$type: "application/json;vnd.sage=syracuse",
							$url: "{$url}",
							$method: "{$method}"
						}
					}
				}
			}
		}
	};
	return {
		$type: "application/x-vignette",
		$format: "$page",
		$prototype: $prototype
	};
}

function _addRecordText(record, $bind, css) {
	var div = document.createElement("div");
	div.className = css;
	var text = syra_site.expressionMaker.parse(record, record.dataset[$bind], record.dataset);
	div.innerHTML = syra_site.dom.formatHTMLText(text.replace(/<em>/g, "<em class='" + css + "-em'>"));
	record.body.appendChild(div);
}

function SearchPage() {}

exports.SearchPage = helpers.defineClass(SearchPage, DesktopPage, {
	loadBox: function(initData) {
		this.isMenuBarDisabled = true;
		this.isBookmarkEnabled = false;
		_$searchLinks.$startWith.$title = syra_local.searchSimilar;
		_$searchLinks.$similar.$title = syra_local.searchStartingWith;
		this._modeBtn = syra_menus.addTextButton(syra_local.searchSimilar, "s-search-mode", "onModeClick");
		this._modeBtn.syraMode = "$startWith";
		this._modeBtn.syraItem = this.id;
		DesktopPage.prototype.loadBox.call(this, initData);
		_facetsBar.load(this, {
			resizeDirection: "right"
		});
		var slot = document.createElement("div");
		slot.className = "s-search-options";
		slot.textContent = syra_local.searchOptionLabel;
		slot.appendChild(this._modeBtn);
		this.header.appendChild(slot);
	},
	onModeClick: function() {
		this._modeBtn.syraMode = this._modeBtn.syraMode == "$startWith" ? "$similar" : "$startWith";
		syra_menus.setButtonTitle(this._modeBtn, _$searchLinks[this._modeBtn.syraMode].$title);
		_sendOneSearchRequest(this, this.dataset, {});
		_sendOneSearchRequest(this, this.dataset, {}, true);
	},
	ensureRepresentation: function() {
		return {
			$prototype: {
				$baseType: "application/json;vnd.sage=syracuse",
				$url: "{$baseUrl}/queryResults",
				$type: "application/json",
				$descriptor: "prototype queryResult.$search",
				$properties: {
					$vignetteFunctions: _getVignetteProto("$queryFunctions"),
					$vignetteData: _getVignetteProto("$queryData"),
					"query": {
						$type: "application/x-string"
					}
				}
			}
		};
	},
	ensureDefaultArticle: function($article) {
		return {
			$title: syra_local.searchPageTitle,
			$layout: {
				$layoutType: "row",
				$items: [{
					$category: "section",
					$title: syra_local.search_functions_tab,
					$clientId: "searchFunctionsTab",
					$layout: {
						$items: [{
							$bind: "$vignetteFunctions"
						}]
					}
				}, {
					$category: "section",
					$clientId: "searchDataTab",
					$title: syra_local.search_data_tab,
					$layout: {
						$items: [{
							$bind: "$vignetteData"
						}]
					}
				}]
			}
		};
	},
	onFieldInputEvent: function(event) {
		_facetsBar.onFieldInputEvent(event);
	},
	onItemInOut: function(onEnter, event, target) {
		_facetsBar.onItemInOut(onEnter, event, target);
	},
	applyChange: function(newData) {
		DesktopPage.prototype.applyChange.call(this, newData);
		newData && newData.$searchFacets && _facetsBar.fill(this.dataset.$searchFacets);
	},
	dispose: function() {
		_facetsBar.dispose();
		DesktopPage.prototype.dispose.call(this);
	},
	fetch: function(options) {
		this.dataset.$search = this.openerUrlSegments.params.q;
		if (!options || Object.keys(options).length == 0) {
			_sendAllSearchRequests(this, this.dataset, {});
		} else {
			// filtering /!\ on data result only
			if (options.jsonParams) {
				_sendOneSearchRequest(this, this.dataset, options, true);
			}
			// pager step navigation (pagination)
			else {
				if (options.$location) {
					_sendOneSearchRequest(this, this.dataset, options, options.$location.$url.indexOf("source=data") != -1);
				}
				// others (simple pager navigation) (pagination)
				// get corresponding tab information, update content accordingly
				else {
					if (options.field && options.field.$item && options.field.$item.$bind) {
						options.jsonParams = this.jsonParams;
						_sendOneSearchRequest(this, this.dataset, options, options.field.$item.$bind == "$queryData");
					} else {
						// in order to handle navigation issues (prev/next) issue #2824
						if (this.openerUrlSegments.params.source) {
							options.params = {
								count: this.openerUrlSegments.params.count,
								startIndex: this.openerUrlSegments.params.startIndex
							};
						}
						_sendAllSearchRequests(this, this.dataset, options);
					}
				}
			}
		}
	},
	setFilterQueryParam: function(options) {
		this.jsonParams = options.jsonParams;
	}
});
