// syracuse-tablet/selenium-tests/test/lib/selectors.js
"use strict";

/**
 *
 */
exports.$$Id = function(id) {
	return {
		id: id
	};
};

/**
 *
 */
exports.$$Action = function(action, params) {
	action = '[data-action="' + action + '"]';
	params = (params && '[data-params="' + params + '"]') || "";

	return {
		css: action + params
	};
};

/**
 *
 */
exports.$$Attr = function(attr, value) {
	if (value != null) {
		attr = '[' + attr + '="' + value + '"]';
	} else {
		attr = '[' + attr + ']';
	}

	return {
		css: attr
	};
};

/**
 *
 */
exports.$$Class = function(cls) {
	return {
		className: cls
	};
};

/**
 * Should not be used in test scripts only in functions that wrap a logical element
 */
exports.$$Css = function(css) {
	return {
		css: css
	};
};

/**
 *
 */
exports.$$Page = function(pageName) {
	return {
		css: '[data-s-ut-page-name="' + pageName + '"]' + '[data-s-ut-page-is-current="' + pageName + '"]'
	};
};

/**
 *
 */
exports.$$Modal = function(modalName) {
	return {
		id: modalName
	};
};

/**
 *
 */
exports.isModal = function(webElement, modalName) {
	return webElement.getAttribute("id").then(function(value) {
		return value == modalName;
	});
};
