// syracuse-tablet/html/js/controls/ctrlCubeChart.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("ctrl");
var Base = require('syracuse-tablet/html/js/controls/ctrlCubeChartBase').Klass;
var uiUtils = require('syracuse-tablet/html/js/ui/uiUtils');
var sdataUtils = require('syracuse-tablet/html/js/sdata/sdataUtils');
var formatApi = require('syracuse-tablet/html/js/helpers/formatApi');
var locale = require("syracuse-tablet/html/js/helpers/locale");

/**
 * cube chart control class
 */
var _Klass = utils.defineClass(function(controller, article, prototype) {
	var self = this;
	Base.call(self, controller, article, prototype);
	self.itemProto = prototype.getPrototype("$item");
}, Base, {
	/**
	 * Build Html by using a builder class
	 */
	buildHtml: function(parent, controllerDao, buildOptions) {
		var self = this;

		self.container = uiUtils.createDomElement('div', ["s-m-control", "s-m-chart"], null, {
			"id": this.id
		});

		self.chartSlot = uiUtils.createDomElement('div', ["s-m-control", "s-m-chart-slot"], null, null, self.container);

		self._buildChart();
		parent.appendChild(self.container);
	},

	_buildChart: function() {
		var self = this;

		// This block assignes some variables so that the implementations
		// in ctrlCubeChartBase which have been migrated from the desktop client
		// do not need to be modified too much!
		self.$prototype = self.prototype.json;
		self.$prototype.$properties = self.$prototype.$item.$properties;
		self._localize = {
			highCharts_showHideLegend: locale.text("highCharts_showHideLegend"),
			highCharts_seriesTitle: locale.text("highCharts_seriesTitle")
		};
		//

		self._initializeMeta();

		self.dataset = self.controller.dao.getQueryDataRaw();

		var settings = self._settingsFromValue();
		self._createChart(settings);
	},


	_ensureFormatter: function(measure) {
		measure.formatApi = {
			format: function(val, localFormat) {
				return val;
			}
		};
		//		measure.formatObj = measure.formatObj || formatApi.getLocale().getNumberFormatObj(measure.$type);
		//		measure.localeFormat = measure.localeFormat || measure.$format || measure.formatObj.numFormat;
	},

	_localizedText: function(key) {
		return key;
	},

	_getFieldEvalTitle: function() {
		return [];
	},

	destroy: function() {
		var self = this;
		Base.prototype.destroy.call(self);
		self.dataset = null;
	}
});

exports.Klass = _Klass;
