// syracuse-tablet/html/js/controls/ctrlPageHeader.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var uiUtils = require('syracuse-tablet/html/js/ui/uiUtils');
var templates = require('syracuse-tablet/html/js/helpers/templating');
var brandNameLabel = "Sage ERP X3 - Mobile";
var logoutLabel = "Logout";

/**
 * Page header Generate Html
 */
var _Klass = utils.defineClass(

	function(page, prototype, $article) {
		var self = this;
		self.prototype = prototype;
		self.$article = $article;
		self.page = page;
	}, null, {

		destroy: function() {
			var self = this;
			self.prototype = null;
			self.$article = null;
			self.page = null;
			if (!self.$elmt) return;
			utils.unbindObj(self.$elmt);
			self.$elmt = null;
			utils.unbindObj(self);
		},

		loadStructure: function() {

		},

		buildHtml: function(parent, controllerDao, buildOptions) {
			var self = this;
			var header = uiUtils.createDomElement("nav", ["navbar", "navbar-inverse"], null, {
				"role": "navigation"
			});

			var containerFluid = uiUtils.createDomElement("div", ["container-fluid"], null, null, header);
			var navbar = uiUtils.createDomElement("div", ["navbar-header"], null, null, containerFluid);
			var picker = uiUtils.createDomElement("button", ["navbar-toggle"], null, {
				"data-toggle": "collapse",
				"data-target": "#header-collapse-" + this.page.id
			}, navbar);
			// append icon bar span (3 times)
			for (var ii = 0, jj = 3; ii < jj; ii++) {
				uiUtils.createDomElement("span", ["icon-bar"], null, null, picker);
			}
			uiUtils.createDomElement("a", ["navbar-brand"], brandNameLabel, {
				"href": "#"
			}, navbar);

			// building list of header items
			var navbarCollapse = uiUtils.createDomElement("div", ["collapse", "navbar-collapse"], null, {
				"id": "header-collapse-" + this.page.id
			}, containerFluid);
			var listContainer = uiUtils.createDomElement("ul", ["nav", "navbar-nav", "navbar-right"], null, null, navbarCollapse);
			var listItem = uiUtils.createDomElement("li", null, null, null, listContainer);
			var anchor = uiUtils.createDomElement("a", null, logoutLabel, {
				"href": "#",
				"data-action": "logout"
			}, listItem);
			uiUtils.createDomElement("span", ["glyphicon", "glyphicon-log-out"], null, null, anchor);

			parent.appendChild(header);
		}
		//,
		/**
		 *	Build html using description json and buildDom() from uiUtils
		 **/
		/*
	xbuildHtml:function(parent, controllerDao, buildOptions){
		var headerDesc = {
			nav: {
				classList: ["navbar", "navbar-inverse"],
				attr: {
					"role": "navigation"
				},
				children: [{
					div: {
						classList: ["container-fluid"],
						children: [{
							div: {
								classList: ["navbar-header"],
								children: [{
									button: {
										classList: ["navbar-toggle"],
										attr: {
											"type": "button",
											"data-toggle": "collapse",
											"data-target": "#header-collapse-" + this.page.id
										},
										children: [{
											span: {
												classList: ["icon-bar"]
											}
										}, {
											span: {
												classList: ["icon-bar"]
											}
										}, {
											span: {
												classList: ["icon-bar"]
											}
										}]
									}
								}, {
									a: {
										classList: ["navbar-brand"],
										textContent: brandNameLabel,
										attr: {
											"href": "#"
										}
									}
								}]
							}
						}, {
							div: {
								classList: ["collapse", "navbar-collapse"],
								attr: {
									"id": "header-collapse-" + this.page.id
								},
								children: [{
									ul: {
										classList: ["nav", "navbar-nav", "navbar-right"],
										children: [{
											li: {
												children: [{
													a: {
														attr: {
															"href": "#",
															"data-action": "logout"
														},
														textContent: logoutLabel,
														children: [{
															span: {
																classList: ["glyphicon", "glyphicon-log-out"]
															}
														}]
													}
												}]
											}
										}]
									}
								}]
							}
						}]
					}
				}]
			}
		};

		uiUtils.buildDom(headerDesc, parent);
	}*/
	});

exports.Klass = _Klass;
