// syracuse-tablet/html/js/controls/ctrlCombo.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var uiUtils = require('syracuse-tablet/html/js/ui/uiUtils');
var Base = require('syracuse-tablet/html/js/controls/ctrlFieldBase').Klass;

var _Klass = utils.defineClass(

	function(controller, article, prototype) {
		var self = this;
		Base.call(self, controller, article, prototype);
	}, Base, {
		buildFieldValue: function(parentSlot, buildOptions) {
			var self = this;
			self.choiceList = this.prototype.data("$value").$enum;
			if (self.$isEditMode) {
				var $format = this.prototype.data("$format");

				// if '$format' is set it can only have values '$combo' or '$radio'
				if ($format && ($format != '$radios' && $format != '$combo')) {
					throw new Error("Choice can only have formats '$radios' or '$combo'");
				}

				// ## set display in case of edit mode ##
				var value = self.currentValue = self.getValue();
				var domValue;

				if (self.choiceList.length > 0) {
					// display in case of '$radios' format
					if (self.prototype.data("$format") == '$radios') {
						self.radioItemList = [];
						$.each(self.choiceList, function(index, choiceItem) {
							// building choice item display
							var choiceItemDom = uiUtils.createDomElement("div", ["sm-field-radio-item"]);
							var inputDom = uiUtils.createDomElement("input", ["sm-field-choice-radio"], null, {
								"type": "radio",
								"value": index,
								"name": "sm-radio-" + self.id,
								"data-ctrl-id": self.id
							});
							if (value && choiceItem.$value == value) {
								inputDom.setAttribute("checked", true);
							}
							var titleDom = uiUtils.createDomElement("div", ["sm-field-choice-radio-title"], choiceItem.$title);

							// build radioItemList array
							self.radioItemList.push({
								"inputDom": inputDom,
								"titleDom": titleDom
							});

							// appending to parent layout
							choiceItemDom.appendChild(inputDom);
							choiceItemDom.appendChild(titleDom);
							parentSlot.appendChild(choiceItemDom);
						});
					}

					// default format is '$combo'
					else {
						// building choice item display with bootstrap btn-group component
						var btnGroup = uiUtils.createDomElement("div", ["btn-group"]);
						var btnTxt = value ? this._getChoiceTitle(value) : self.choiceList[0].$title;
						self.selectionBtn = uiUtils.createDomElement("button", ["btn", "btn-default", "dropdown-toggle"], btnTxt, {
							"type": "button",
							"data-toggle": "dropdown"
						});
						// [style] add caret to selection button
						self.selectionBtn.appendChild(uiUtils.createDomElement("span", ["caret"]));

						self.listContainer = uiUtils.createDomElement("ul", ["dropdown-menu"], null, {
							"role": "menu"
						});
						// building list
						$.each(self.choiceList, function(index, choiceItem) {
							var listItem = uiUtils.createDomElement("li", (choiceItem.$value == value) || (!value && index == 0) ? ["active"] : null);
							listItem.appendChild(uiUtils.createDomElement("a", ["sm-field-choice-combo-item"], choiceItem.$title, {
								"href": "#",
								"data-ctrl-id": self.id
							}));

							// appending list to list container
							self.listContainer.appendChild(listItem);
						});

						// appending to parent layout
						btnGroup.appendChild(self.selectionBtn);
						btnGroup.appendChild(self.listContainer);
						parentSlot.appendChild(btnGroup);
					}
				}
			} else {
				if (self.choiceList.length > 0) {
					parentSlot.appendChild(self.fieldValue = uiUtils.createDomElement("div", ["sm-field-value-read"], self._getChoiceTitle(self.getValue())));
				}
			}
		},
		_getChoiceTitle: function(value) {
			var self = this;
			var res;
			$.each(self.choiceList, function(index, choice) {
				res = choice.$value == value ? choice.$title : res;
				if (res) {
					return false;
				}
			});
			return res;
		},
		_getChoiceValue: function(title) {
			var self = this;
			var res;
			$.each(self.choiceList, function(index, choice) {
				res = choice.$title == title ? choice.$value : res;
				if (res) {
					return false;
				}
			});
			return res;
		},
		setDisplayValue: function(value, index, pickerText) {
			var self = this;

			if (!self.$isEditMode) {
				self.fieldValue.textContent = self._getChoiceTitle(value);
			} else {
				// in case of radio choice
				if (self.prototype.data("$format") == '$radios') {
					// apply change if new value
					if (value != self.currentValue) {
						// update widget
						$.each(self.radioItemList, function(idx, radioItem) {
							setTimeout(function() {
								// if new choice index is defined
								if (index) {
									radioItem.inputDom.checked = index == idx;
								}
								// if new choice index is not defined
								else {
									radioItem.inputDom.checked = self.choiceList[idx].$value == value;
								}
							}, 5);
						});
					}
				}


				// in case of combo choice
				else {
					// apply change if new value
					if (value != self.currentValue) {
						// update widget
						$.each(self.listContainer.children, function(idx, listItem) {
							// if new choice index is defined
							if (index) {
								uiUtils.toggleClass(listItem, "active", index == idx);
							}
							// if new choice index is not defined
							else {
								uiUtils.toggleClass(listItem, "active", self.choiceList[idx].$value == value);
							}
						});
						self.selectionBtn.textContent = pickerText || self._getChoiceTitle(value);
						self.selectionBtn.appendChild(uiUtils.createDomElement("span", ["caret"]));
					}
				}

			}
		},
		setFieldValue: function(value, index, pickerText) {
			var self = this;

			self.setDisplayValue(value, index, pickerText);

			self.currentValue = value;
			self.dao.setValue(self.article.$bind, value);
		},
		onClick: function(evt) {
			var self = this;
			if (self.prototype.data("$format") == '$radios') {
				self.setFieldValue(self._getChoiceValue(evt.target.nextElementSibling.textContent), $(evt.target.parentNode).index());
			} else {
				self.setFieldValue(self._getChoiceValue(evt.target.textContent), $(evt.target.parentNode).index(), evt.target.textContent);
				// close combo selector
				$(self.selectionBtn).dropdown('toggle');
			}
		}
	});

exports.Klass = _Klass;
