// syracuse-tablet/html/js/controls/hammerScroller.js
"use strict";

var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("scroll", false);
var utils = require('syracuse-tablet/html/js/helpers/utils');
var globals = require('syracuse-tablet/html/js/helpers/globals');
var uiRect = require('syracuse-tablet/html/js/ui/rect');
var jsutils = require('syracuse-tablet/html/js/helpers/jsutils');
var notifications = require('syracuse-tablet/html/js/helpers/notifications');

var _notifiedNoScroll = false;

notifications.subscribe(function(message, status) {
	// Save global status
	status = status == null ? !_notifiedNoScroll : status === true;
	_notifiedNoScroll = status;
	// Toggles Disable/Enable all scrollers
	notifications.publish("sm.scroller.toggle", !_notifiedNoScroll);
}, ["sm.hammer.scrolling.toggle"], 1);

var _stopClickEvent = function(mgr, evt, reason) {
	log && log("Clicked canceled", reason, mgr.id);
	evt.stopPropagation();
	evt.preventDefault();
	evt.$stop = true;
};
// Time to wait before allowing click events a scrolling
var _lastClickTO = 400;
// Bottom/right margin  - we scroll _margin pixels more that the height/width of viewRect
var _margin = 0;
// Enable the slowdown process - scroll continue after panend (mouse up ) like on native devices
var _slowDown = true;
// Move the $$elmt after _slowDownTO in slowdown process
var _slowDownTO = 30;
/**
 * Creates a vertical/horizontal scroller which emulates the tablet native behavior
 * 		$$elmt		element to scroll
 * 		options		callBack		Optional Called on event and skip default process
 * 					valMax			maximum top - 0 by default - 50 for pages with header...
 * 					direction		h/v
 * 					isPageScroller	true/false
 * 		parentMgr	parent scroll manager (the page's one for an array scroller)
 * 					used to deal with click event
 * 		$$nativeScroll 	div that will take the native scroll when we disable the scrolling (for test and authoring)
 * 						optional - $elmt if null
 */
var _consts = {
	pan: {
		v: "panup pandown panstart panend",
		h: "panleft panright panstart panend"
	},
	dir: {
		v: Hammer.DIRECTION_VERTICAL,
		h: Hammer.DIRECTION_HORIZONTAL
	},
	upEvt: {
		v: "panup",
		h: "panleft"
	}
};
var _Klass = utils.defineClass(function hammerScroller($$elmt, options, parentMgr, $$nativeScroll) {
	options = options || {};
	this.hammer = null;
	this.$$elmt = $$elmt;
	this.$$nativeScroll = $$nativeScroll || $$elmt;
	if ($$elmt == "null") throw new Error("hammerScroller - null $$elmt");
	this.id = $$elmt.closest("[id]").attr("id");
	// Max top always 0 because we scroll in relative / parent - If needed pass it in parameter
	this.valMax = options.valMax != null ? options.valMax : 0;
	this.name = options.name || "scroller";
	this.parentMgr = parentMgr;
	this.callBack = options.callBack;
	this.isPageScroller = options.isPageScroller === true;
	// Init hammer manager
	this.hammer = new Hammer(this.$$elmt.get(0));
	// Direction of scroller
	if (options.direction) {
		this._initDir(options.direction);
	}
	// evts enabled - false by default
	this._evtsEnabled = false;
	// To allow disable/Enable - General HammerScroller objects - all object will be toggled
	notifications.subscribe(this, ["sm.scroller.toggle"], 1);
}, null, {
	/**
	 * Init direction and associated pan events
	 */
	_initDir: function(direction) {
		this.dir = direction;
		if (this.dir !== "v" && this.dir !== "h") throw new Error("hammerScroller - Unexpected direction [" + this.dir + "]");
		// event than means up or right
		this.upEvtType = _consts.upEvt[this.dir];
	},
	/**
	 * We need to initialize the scroll after creation
	 * Scrollable object arrays/pages have an empty scroll which is initialized if needed
	 * Initialization needs to be called when the object has been displayed (after show) to make sure that viewRect is ok
	 * To manage savedCtxRestore it's also the better way
	 * 		- save ctx (top)is stored and took into account at initialization
	 * Params
	 * 		viewRect		rectangle (uiUtils/rect class) of the view within we will move the content of $$elmt (kind of window)
	 * 						!!!top/left must be in fixed coordinates (relative to document)
	 * 						in order to be able to check if mouse pointer belongs to this rect on mousemove/pandown
	 * 		dir				To change dir if needed (orientation changed)
	 * 		valMax			To change valMax if needed (orientation changed)
	 */
	init: function(viewRect, dir, valMax) {
		if (viewRect == null) throw new Error("hammerScroller - Unexpected null viewRect");
		// Reset the scroller - not all values - keep valMax, dir, restoredVal...
		this.reset();
		// Init can override direction
		if (dir != null && this.dir != dir) {
			this._initDir(dir);
		}
		if (this.dir == null) throw new Error("hammerScroller - Unexpected null direction");
		// viewRect defines the boundaries
		this.viewRect = viewRect;
		// We need pan and not swipe
		this.hammer.get("pan").set({
			threshold: 1, // Small threshold to bloc the click (mouseup) as son as pan start- To test on mobile devices
			direction: _consts.dir[this.dir]
		});
		// We need press to stop the scolling
		this.hammer.get("press").set({
			time: 1 //Small timeout needed to stop immediately when press
		});
		// Create handlers
		this._panHandler = Hammer.bindFn(this.onPan, this);
		this._pressHandler = Hammer.bindFn(this.onPress, this);
		this._clickHandler = Hammer.bindFn(this.onClick, this);
		// - altClick enable/disable the scrolling
		this.$$elmt.on('click', this._clickHandler);
		// Bind events
		// Original rectangle - used to enable/disable scrolling
		// !!! Images and components loaded by URL need to have a fix height !!! otherwise we can't calculate the height
		this._originalRect = uiRect.elmtRect(this.$$elmt);
		this._enableEvents(true);
		// Init can override valMax
		if (valMax != null) this.valMax = valMax;
		// Values set by restoreSaveContext - back 
		this.val = this.restoredVal == null ? this.valMax : this.restoredVal;
		this.restoredVal = null;
		this.refresh();
	},
	/**
	 * Enable / Disable events
	 * Used by reset and Alt Click
	 */
	_enableEvents: function(yes) {
		//this.viewRect == null -> scroller has not been initialized
		// Can occurs because we create a non-initialized scroller for pages/array so they can be notified even if they've not been initialized
		if (!this.viewRect) return;
		// Object has been notified to not enable scroll
		yes = yes && !_notifiedNoScroll;
		var wasDisabled = this._evtsEnabled === false;
		// ENABLED
		if (yes === true && this._evtsEnabled !== true) {
			this._evtsEnabled = true;
			this.hammer.on(_consts.pan[this.dir], this._panHandler);
			this.hammer.on("press", this._pressHandler);
			// Just to find elmt with debugger
			this.$$elmt.addClass("s-m-scroll" + this.dir);
		}
		// DISABLED
		if (yes === false && this._evtsEnabled !== false) {
			if (this._evtsEnabled === false) return;
			this._evtsEnabled = false;
			if (this.hammer) {
				this.hammer.off(_consts.pan[this.dir], this._panHandler);
				this.hammer.off("press", this._pressHandler);
			}
			this.$$elmt.removeClass("s-m-scroll" + this.dir);
		}
		// IN ANY CASES
		if (wasDisabled) {
			// Nothing
		} else {
			this.panStart = false;
			this.stopSlowingDown();
			this.cancelClick = false;
			this.lastClickAllowed = null;
		}
		/**
		 * Set $$elmt and $$nativeScroll to enable/disable native scrolling (browser's scroll bar)
		 */
		// Set $$nativeScroll properties - overflow and width/height
		var css = {
			overflow: _notifiedNoScroll === true ? "auto" : "hidden"
		};
		// Set the width/height with right values
		// _notifiedNoScroll = true -> We force the display of natve scrolls otherwise
		var prop = this.dir === "h" ? "width" : "height";
		css[prop] = (_notifiedNoScroll === true ? this.viewRect[prop] : this._originalRect[prop]) + "px";
		this.$$nativeScroll.css(css);
		// Set $$elmts properties - top and overflow
		var css = {
			overflow: _notifiedNoScroll === true ? "auto" : "hidden"
		};
		// Position to default offset
		css[this.dir === "h" ? "left" : "top"] = this.valMax + "px";
		this.$$elmt.css(css);
		this.val = this.valMax;
	},
	/**
	 * Disable scroll but keep the capability to re-initialize it
	 */
	reset: function() {
		// !! FIRST - Disable events
		this._enableEvents(false);
		// Disable click - No mo more scrolling compared to disable/enable
		this.$$elmt.off('click', this._clickHandler);
		// After
		this.viewRect = null;
		// True if panstart - mouse start to move
		this.panStart = false;
		// True if the scroll is slowing down after panend (mouse up)
		this.stopSlowingDown();
		// Cancel click on a link
		this.cancelClick = false;
		//Prevent click after _lastClickTO ms after a panend
		this.lastClickAllowed = null;
		// Null to force to set value on init ( context restored)
		this.val = null;
		// Null to force to set value on pan event
		this.valMin = null;
		this.isLocked = false;
		// !! DO NOT RESET - dir, valMax, parentMgr, restoredVal
	},
	/**
	 * Destroy scroller
	 */
	destroy: function() {
		// First
		this.reset();
		// After
		if (this.hammer) {
			this.hammer.destroy();
			this.hammer = null;
		}
		this.$$elmt = null;
		this.$$nativeScroll = null;
		this.callBack = null;
		this._panHandler = null;
		this._pressHandler = null;
		this._toggleRect(false);
		this._clickHandler = null;
		utils.unbindObj(this);
	},
	/**
	 * Refresh scrolling
	 * 	internal call or called to refresh scrolling when we activate cached pages
	 */
	refresh: function() {
		if (this.dir == "v") {
			this.$$elmt.css({
				top: this.val + "px"
			});
		} else {
			this.$$elmt.css({
				left: this.val + "px"
			});
		}
	},
	_toggleRect: function(remove) {
		// Debug show the rectangle - fixed position 
		if (this._$$viewRect) {
			this._$$viewRect.remove();
			if (this._$$docRect) this._$$docRect.remove();
			this._$$docRect = null;
			this._$$viewRect.off('click', this._clickHandler);
			this._$$viewRect = null;
		} else if (remove !== false) {
			this._$$viewRect = this.viewRect.showInDocument();
			var o = this.isPageScroller ? {
				top: 0,
				left: 0
			} : globals.getCurrentPage().getScrollOffset();
			this._$$viewRect.css({
				marginTop: "+=" + o.top,
				marginLeft: "+=" + o.left
			});
			this._$$docRect = uiRect.elmtRect(globals.getCurrentPage().$$elmt).showInDocument({
				color: "green",
				zIndex: 0
			});
			this._$$viewRect.on('click', this._clickHandler);
		}
	},
	/**
	 * Intercept click/mouseup events for filtering
	 * We need to filter click events to control the triggering of links
	 * altClick 	-> Enables/Disables scrolling
	 * shiftClick	-> Displays/Hides the viewRect
	 */
	onClick: function(evt) {
		log && log("onClick Status", this.slowingDown, this.panStart, this.cancelClick, this.lastClickAllowed ? new Date().getTime() - this.lastClickAllowed : null);
		evt.$stop = false;
		if (evt.altKey) {
			// Toggles Disable/Enable hammer scrolling global setting
			notifications.publish("sm.hammer.scrolling.toggle");
			return _stopClickEvent(this, evt, "Disable/Enable events");
		}
		if (evt.shiftKey) {
			this._toggleRect();
			return _stopClickEvent(this, evt, "Show rectangle");
		}
		// If disable we skip
		if (this._evtsEnabled === false) return;
		if (this.slowingDown) {
			return _stopClickEvent(this, evt, "slowing down");
		}
		if (this.cancelClick) {
			this.cancelClick = false;
			return _stopClickEvent(this, evt, "Click canceled");
		}
		if (this.lastClickAllowed != null && ((new Date().getTime() - this.lastClickAllowed) < 0)) {
			// Click is authorized if not slowingDown or if not cancelClick
			// We wait _lastClickTO after panend (mouseup) to allow click
			this.lastClickAllowed = null;
			return _stopClickEvent(this, evt, "Last click timeout");
		}
		if (this.parentMgr && this.parentMgr.dir != this.dir) {
			// For double scroll direction one by parent and one by a grid we need to sak the parent to cancel the click 
			this.parentMgr.onClick(evt);
			if (evt.$stop === true) {
				return _stopClickEvent(this, evt, "Parent stopped click");
			}
		}
	},
	/**
	 * Used to stop the slowdown process
	 */
	onPress: function(evt) {
		this.lastPress = new Date().getTime();
		if (this.slowingDown) {
			this.stopSlowingDown();
			// Last click is avoided - The click evt that is always triggered after press a event
			this.cancelClick = true;
			// Wait _lastClickTO to avoid clik on mouseup
			//this.lastClickAllowed = new Date().getTime() + _lastClickTO;
		}
		if (this.parentMgr) this.parentMgr.isLocked = false;
		log && log(this.id, "onPress", "slowingDown=" + this.slowingDown, "panStart=" + this.panStart, "cancelClick =" + this.cancelClick, "lastClickAllowed =" + this.lastClickAllowed);
	},
	/**
	 * pan event - mousedown/mousemove/mouseup
	 */
	onPan: function(evt) {
		log && log(this.id, evt.type, this.dir, utils.isEvtDirection(evt, this.dir), "isLocked", this.isLocked);
		if (!utils.isEvtDirection(evt, this.dir)) return;
		if (this.isLocked === true) return;
		try {
			if (this.valMin == null) {
				// We initialize this.val here because I noticed that $$elmt.height could be wrong if we in constructor
				// To check again - Maybe it comes from image that are loaded async - We need to set fix width/height for images		
				var prop = this.dir == "v" ? "height" : "width";
				this.valMin = this.valMax - this.$$elmt[prop]() - _margin + this.viewRect[prop];
			}
			if (!this.viewRect.isEvtInside(evt.srcEvent, this.isPageScroller ? null : globals.getCurrentPage().getScrollOffset())) {
				log && log(globals.getCurrentPage().getScrollOffset());
				// outside the viewRect -> Stop process
				log && log("outside");
				return;
			}
			if (evt.type === "panstart") {
				// Mouse mousedown + mousemove event - start the process
				log && log(evt.type);
				this.panStart = true;
				this.cancelClick = true;
				this.stopSlowingDown();
				if (this.parentMgr && (this.parentMgr.dir === this.dir)) this.parentMgr.isLocked = true;
			} else if (evt.type === "panend") {
				// mouseup  - we continue to scroll (_slowDown) until the user do a press event to stop scrolling
				log && log(evt.type);
				this.panStart = false;
				if (_slowDown && this.lastVelocity != null) {
					// We keep the last velocity for slowdown process
					this.lastVelocity = Math.abs(this.lastVelocity);
					if (this.lastVelocity > 0.3 && this.lastVelocity < 1) {
						// If velocity is too small we keep 1
						this.lastVelocity = 1;
					} else {
						// under.3 we consider that we need to stop scrolling
						this.lastVelocity = Math.min(this.lastVelocity, 2);
					}
					if (this.lastVelocity >= 1) {
						// _slowDown option continue the process by decreasing the speed slowly and moving $$elmt
						this.slowingDown = true;
						this.slowDown(this.lastVelocity, this.lastUp);
						this.lastVelocity = this.lastUp = null;
					} else {
						if (this.parentMgr) this.parentMgr.isLocked = false;
					}
				} else {
					// avoid click on mouse up
					this.lastClickAllowed = new Date().getTime() + _lastClickTO;
					if (this.parentMgr) this.parentMgr.isLocked = false;
				}
				return;
			}
			// Direction of the scroll
			// Up means up or right 
			var up = evt.type === this.upEvtType || evt.type === "panstart" && evt.deltaY < 0;
			if (this.move(evt.velocity, up)) {
				this.lastVelocity = Math.abs(evt.velocity);
				this.lastUp = up;
			} else {
				this.lastVelocity = this.lastUp = null;
			}
		} catch (e) {
			log && log(JSON.stringify(jsutils.convertToDiagnoses("hammer scroller error", e), null, 2));
			this.destroy();
			return false;
		}
	},
	stopSlowingDown: function() {
		this.slowingDown = false;
		if (this.parentMgr) this.parentMgr.isLocked = false;
	},
	/**
	 * Keep the last speed to continue scrolling process until the user do a press evt
	 */
	slowDown: function(v, up) {
		var self = this;
		if (self.panStart || !self.slowingDown) {
			this.stopSlowingDown();
			return;
		}
		// keep the same velocity - to improve - we can decrease...
		if (v < 1) {
			this.stopSlowingDown();
			return;
		}
		if (self.move(v, up)) {
			// Move until stop with _slowDownTO timeout
			setTimeout(function() {
				self.slowDown(v, up);
			}, _slowDownTO);
		} else {
			this.stopSlowingDown();
		}
	},
	/**
	 * Move $$elmt vertically
	 */
	move: function(velocity, up) {
		try {
			// Velocity of the move - 0 if we move the finger slowly and ~ 2 for fast
			var v = Math.abs(Math.round(velocity * 5));
			// Distance to move - square function of velocity (non linear) 
			// Allow to move slow and fast if we increase the speed of the finger
			var dist = Math.max(1, v * v / 2);
			if (dist == 0) return;
			// Check boundaries
			var val = up ? this.val - dist : this.val + dist;
			val = Math.max(Math.min(this.valMax, val), this.valMin);
			if (val === this.valMax || val === this.valMin) {
				log && log(this.id, "end scrolling");
				// avoid click during xx ms on mouse up at the end of scrolling
				this.lastClickAllowed = new Date().getTime() + _lastClickTO;
				this.cancelClick = false;
				if (this.val === val) {
					// Overflow -> Stop
					// log && log("move overflow")
					return false;
				}
			}
			this.val = val;
			if (this.callBack) {
				return this.callBack(this.val);
			} else {
				// Move elmt
				this.refresh();
				return true;
			}
		} catch (e) {
			log && log("Page VerticalScroller onMove Error - disable scroll");
			log && log(JSON.stringify(jsutils.convertToDiagnoses("VScrolling error", e), null, 2));
			this.destroy();
			return false;
		}
	},
	// Save context if enabled
	savedCtxCreate: function() {
		return this.hammer ? {
			val: this.val
		} : null;
	},
	// Restored context
	savedCtxRestore: function(ctx, refresh) {
		if (ctx == null || ctx.val == null) return;
		// Set val - if refresh == false that means that we need to call init method 
		this.restoredVal = ctx.val;
		if (refresh) {
			// By example called by a cached pages which has been activated
			this.refresh();
		}
	},
	getOffset: function() {
		return {
			top: this.dir == "v" ? this.val : 0,
			left: this.dir == "h" ? this.val : 0
		};
	},
	/**
	 * status	null	 	toggles this._evtsEnabled otherwise
	 * 			true 		forces enable
	 * 			false		forces disable
	 */
	notifScrollerToggle: function(status) {
		this._enableEvents(status == null ? !this._evtsEnabled : status === true);
	}
});

exports.Klass = _Klass;
