// syracuse-tablet/html/js/controls/layoutStack.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var Base = require('syracuse-tablet/html/js/controls/layoutBase').Klass;

/**
 * cssOptions
 * 		cssLayout, cssCell: class added to default ones
 * default classes can  be overridden
 */
var _Klass = utils.defineClass(
	function(controller, type, article) {
		var self = this;
		Base.call(self, controller, type, article);
	}, Base, {
		defaultDisplayOptions: function() {
			return {
				cssDefLayout: ["container s-m-layout s-m-stack"],
				cssDefCell: ["s-m-layout-cell"]
			};
		},
		buildHtmlBeforeChild: function(html, controllerDao, child, idx) {
			var self = this;
			html.push('<div class="');
			html.push(self.displayOptions.cssDefCell.join(' '));
			if (self.displayOptions.cssCell) {
				html.push(' ');
				html.push(self.displayOptions.cssCell.join(' '));
			}
			html.push('">');
		},

		buildHtmlAfterChild: function(html, controllerDao, child, idx) {
			var self = this;
			html.push('</div>');
		},

		buildHtmlBefore: function(html, controllerDao) {
			var self = this;
			html.push('<div class="');
			html.push(self.displayOptions.cssDefLayout.join(' '));
			if (self.displayOptions.cssLayout) {
				html.push(' ');
				html.push(self.displayOptions.cssLayout.join(' '));
			}
			html.push('" id="');
			html.push(self.id);
			html.push('">');
		},

		buildHtmlAfter: function(html, controllerDao) {
			var self = this;
			html.push('</div>');
		}
	});

exports.Klass = _Klass;
