// syracuse-tablet/html/js/controls/validator.js
"use strict";

var locale = require('syracuse-tablet/html/js/helpers/locale');

/*
 *
 * Collection of all SData contraints and a validation function per constraint
 *
 */

var _validator = {
	validate: function(ctrl, value, errors, constraints) {
		var proto = ctrl.prototype;
		var ok = true;
		for (var i = 0; i < constraints.length; i++) {
			var c = constraints[i];
			var prop = proto.data(c);
			if (prop != null) { // constraint set in prototype
				var fn = _validator[c];
				if (fn) {
					ok = ok && fn(value, errors, prop);
				} else {
					errors.push(locale.text("field.unknownConstraint", [c]));
					ok = false;
				}
			}
		}
		return ok;
	},

	// General
	$isMandatory: function(value, errors, constraint) {
		if (constraint && value == null || value == "") {
			errors.push(locale.text("field.valMand"));
			return false;
		}
	},

	// Strings
	$pattern: function(value, errors, constraint) {
		var self = this;
		if (value && constraint) {
			var re = new RegExp(constraint);
			if (!re.test(value)) {
				errors.push(locale.text("field.patternError", [constraint]));
				return false;
			}
		}
		return true;
	},

	// Numeric
	$minimum: function(value, errors, constraint) {
		var self = this;
		if (value != null && value < constraint) {
			errors.push(locale.text("field.minimum", [constraint]));
			return false;
		}
		return true;
	},
	$maximum: function(value, errors, constraint) {
		var self = this;
		if (value != null && value > constraint) {
			errors.push(locale.text("field.maximum", [constraint]));
			return false;
		}
		return true;
	},
	$exclusiveMinimum: function(value, errors, constraint) {
		var self = this;
		if (value != null && value <= constraint) {
			errors.push(locale.text("field.exclusiveMinimum", [constraint]));
			return false;
		}
		return true;
	},
	$exclusiveMaximum: function(value, errors, constraint) {
		var self = this;
		if (value != null && value >= constraint) {
			errors.push(locale.text("field.exclusiveMaximum", [constraint]));
			return false;
		}
		return true;
	},
	$minLength: function(value, errors, constraint) {
		var self = this;
		if (!value || value.length < constraint) {
			errors.push(locale.text("field.minLength", [constraint]));
			return false;
		}
		return true;
	},
	$maxLength: function(value, errors, constraint) {
		var self = this;
		if (value && value.length > constraint) {
			errors.push(locale.text("field.maxLength", [constraint]));
			return false;
		}
		return true;
	},

};

exports.Validator = _validator;
