// syracuse-tablet/html/js/controls/vignetteBase.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var globals = require('syracuse-tablet/html/js/helpers/globals');
var Base = require('syracuse-tablet/html/js/controls/ctrlBase').Klass;
var uiUtils = require('syracuse-tablet/html/js/ui/uiUtils');
var fontUtils = require('syracuse-tablet/html/js/ui/fontUtils');
var PageLoader = require('syracuse-tablet/html/js/application/pageLoader').Klass;
var pageHelper = require('syracuse-tablet/html/js/application/pageHelper');
var modal = require('syracuse-tablet/html/js/ui/modal');
var notifications = require('syracuse-tablet/html/js/helpers/notifications');
var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("app");

var _waitConfig = {
	className: 'waiting-blocks',
	elements: 5,
	auto: true,
	speed: 200
};
var _templates = {
	main: '\
		<section/>\
		<footer style="display:none"/> \
		<div class="s-m-waiting bckg" style="display:none"/> \
		<div class="s-m-waiting wheel" style="display:none"/> \
	',
	iframe: '\
		<br><iframe src="{{url}}" style="position: absolute; top: 0px; left: 0px; height: 100vh; width: 100vh;"/>\
	',
	refreshbtn: '\
		<a href="#" data-action="vignetteRefresh" class="' + fontUtils.vignetteIcons("refresh") + '" style="float:right"></a>\
	',
	navbtns: '\
		<a href="#" data-action="vignetteBack" class="' + fontUtils.vignetteIcons("back") + '" style="display:none"></a>\
		<a href="#" data-action="vignetteHome" class="' + fontUtils.vignetteIcons("home") + '" style="display:none"></a>\
	',
	vignetteError: '\
		<a href="#" data-action="vignetteError">{{text}}</a>\
	',
	linkPage: '<a href="{{href}}" data-nav-target="application" data-nav="{{data-nav}}" data-gadget-id="{{data-gadget-id}}" data-application-name="{{data-application-name}}" class="s-m-link-page">{{title}}</a>'
};

var _getHtml = function(name, ctx) {
	var key = name + "compiled";
	var tmpl = _templates[key];
	if (!tmpl) {
		tmpl = _templates[key] = Handlebars.compile(_templates[name]);
	}
	return tmpl(ctx);
};


var _Klass = utils.defineClass(

	function(controller, article, prototype, opts) {
		var self = this;
		article.$article = article.$article || {};
		Base.call(self, controller, article, prototype, opts);
		self._currentPage = null;
		self._pageLoader = null;
		self._$$wait = null;
		self._stateStack = [];
		self._error = null;
		self.displayStyle = prototype.data("$displayStyle");
		self.openLinkInVignette = globals.$config("openLinkInVignette");
		self.enableRefresh = true;
	}, Base, {
		destroy: function() {
			var self = this;
			self._error = null;
			// destroy all pages attached to the vignette
			self.clearContent();
			uiUtils.waitPluginDestroy(self._$$wait);
			if (self._pageLoader = null) {
				self._pageLoader.destroy();
				self._pageLoader = null;
			}
			self._stateStack = [];
			Base.prototype.destroy.call(self);
		},
		/**
		 * If vignette contains a page return the control object identified by id
		 */
		getPageControl: function(ctrlId) {
			return this._currentPage && this._currentPage.getControl ? this._currentPage.getControl(ctrlId) : null;
		},
		getPage: function() {
			return this._currentPage;
		},
		onResize: function(evt) {
			// We don't propagate parent page resize if page has not been loaded
			// --> onResize will be called when page has been loaded
			// Parent resize event is propagated only if we resize / change display after vignette has been loaded
			if (this._loaded && this._currentPage && this._currentPage.state.options.$displayStyle === "$full") {
				log && log("vignette onResize " + this.id);
				this._currentPage.onResize(evt);
			} else {
				log && log("vignette onResize skipped " + this.id);
			}
		},
		/**
		 * Change current page - called from eventListener
		 * 	data-nav action has been clicked in a vignette's link
		 * 	we could have a parameter that indicates if the link has to be opened full page or inside the vignette
		 */
		changePage: function(state, options) {
			options = options || {};
			// Propagate openLinkInVignette to child page (- used to open reflinks...
			options.openLinkInVignette = this._linkInside();
			// Always full page
			options.$displayStyle = "$full";
			this._loadPage(state, options);
		},
		/**
		 * Refresh current page - called from eventListener
		 */
		refreshPage: function(options) {
			var self = this;
			if (!self._currentPage) throw new Error("No current page");
			self._currentPage.refresh(options, {
				waitStart: function(controlId) {
					// Wait managed by vignette
					self.waitStart();
				},
				waitStop: function(controlId) {
					self.waitStop();
				}
			}).then(function() {
				//	Eventually do something in vignette control	
			}).fail(function(e) {
				modal.error("Refresh page on vignette", e);
			});
		},
		hasBtnsBar: function() {
			return this._linkInside() || this.enableRefresh;
		},
		clearContent: function() {
			var self = this;
			if (self._$$content) {
				pageHelper.removeAllPages(self._$$content);
				self._$$content.empty();
			}
			// _currentPage has ben destroyed by removeAllPages
			if (self._currentPage) self._currentPage = null;
		},
		_linkInside: function(state) {
			if (this.openLinkInVignette === true) return true;
			var name;
			if (state == null) {
				name = this._currentPage ? this._currentPage.state.name : null;
			} else if ($.isPlainObject(state)) {
				name = state.name;
			} else {
				name = state;
			}
			if (name === "rowdetail") this.openLinkInVignette = true;
			if (this._currentPage && this._currentPage.isChild) this.openLinkInVignette = true;;
			return this.openLinkInVignette;
		},
		/**
		 * Load a page
		 * 	action: 	back for a back in vignette actions
		 * 				home for a home in vignette actions
		 */
		_loadPage: function(state, options, action) {
			var self = this;
			self._loaded = false;
			self._error = null;
			if (!self._linkInside(state) && self._stateStack.length > 0) {
				// Clear content only that case. Otherwise we have to store cached paged inside the vignette
				self.clearContent();
			}
			if (!self._pageLoader) {
				/**
				 * Page loader shared with application
				 * Load a page
				 */
				self._pageLoader = new PageLoader({
					waitStop: function() {
						self.waitStop();
					},
					waitStart: function() {
						self.waitStart();
					},
					getRootElmt: function() {
						return self._$$content;
					},
					getCurrentPage: function() {
						return self._currentPage;
					},
					setCurrentPage: function(page) {
						self._currentPage = page;
					},
					historyPush: function(state) {
						// Add state in stack in all case - We need it for link to row Detail
						self._stateStack.push(state);
					},
					succeeded: function(dstPage, parentNotifInfo) {
						if (self.hasBtnsBar()) {
							self.refreshBtnsBar();
						}
						if (parentNotifInfo && parentNotifInfo.notifId) {
							/**
							 * _pageNotify  to notify only the page
							 */
							if (dstPage.id === parentNotifInfo.parentId) {
								notifications.pageNotify(parentNotifInfo.notifId, self._currentPage, parentNotifInfo.notifData, parentNotifInfo.controlId);
							} else {
								modal.error("Child page notification warning", "Parent page is not the current page");
							}
						}
						// page has been loaded - Used to prevent multiple onresize calls to current page
						self._loaded = true;
						log && log("vignette loaded" + self.id);
						// Call resize handler each time a page has been loaded for calculations
						self.onResize(null);
					},
					loadingError: function(deferred, state, e) {
						self._displayError("Loading page [" + state.name + "] error", e);
						// No reject
						deferred.resolve();
					},
					/**
					 * Specific interface method for vignettes
					 * Returns the context of a page nested in a vignette and also context of the vignette (selection)
					 * Called by the pageLoader after the vignette has been displayed
					 */
					savedCtxVignette: function() {
						// It's the dashboard that manage the storage
						return self.controller.savedCtxVignette(self);
					}
				});
			}
			try {
				var pageInfo;
				var articleCopy;
				if (typeof state === "string") {
					options = options || {};
					options.vignette = true;
					articleCopy = options.article; // Backup of original article, this MUST NOT be cloned
					pageInfo = pageHelper.getPageInfo(state, options, self._currentPage);
				} else if (action === "back" || action === "home") {
					pageInfo = state;
				} else {
					throw new Error("Unexpected object state");
				}
				if (articleCopy && pageInfo.options) {
					pageInfo.options.article = articleCopy;
				}
				if (pageInfo.options && pageInfo.options.$displayStyle === "$full") {
					// action "home" is not a back. It's like an open page because we want to store the state in history
					return self._pageLoader.load(pageInfo, action === "back");
				} else if (options.gadget) {
					var deferred = $.Deferred();
					self._addPicker({
						"data-nav": pageInfo.name,
						"data-gadget-id": options.gadget.data("$uuid"),
						"title": options.gadget.data("$title")
					});
					deferred.resolve();
					return deferred.promise();
				} else {
					throw new Error("Unexpected vignette page");
				}
			} catch (e) {
				self._displayError("Loading page error", e);
			}
		},

		_loadRepresentation: function(gadget, name, options) {
			var self = this;
			var name = name || gadget.getPageName();
			options = $.extend({}, options);
			options.openLinkInVignette = self._linkInside(name);
			options.$displayStyle = self.displayStyle;
			options.gadget = gadget;
			options.article = self.article.$article; // sub article in case the layout of a vignette is defined in the dashboard layout
			options.layoutPath = self.getAuthoringPath() + "-0";
			return self._loadPage(name, options);
		},

		_displayError: function(msg, detail) {
			var self = this;
			self._loaded = true;
			self._error = {
				message: msg,
				detail: detail
			};
			self.$$elmt.addClass("s-m-error");
			self.$$elmt.html(_getHtml("vignetteError", {
				text: "Loading failed"
			}));
		},

		buildHtml: function($$parent, controllerDao, buildOptions) {
			var self = this;
			if (self._stateStack.length > 0) {
				// There could be cached pages

			} else {
				Base.prototype.buildHtml.call(self, $$parent, controllerDao, buildOptions, ["s-m-vignette"]);
				self.$$elmt.html(_getHtml("main", {}));
				self._$$content = self.$$elmt.children("section");
				self.$$footer = self.$$elmt.children("footer");
				self._$$wait = uiUtils.waitPluginCreate(self.$$elmt.children(".s-m-waiting.wheel"), _waitConfig);
				self._$$waitBckg = self.$$elmt.children(".s-m-waiting.bckg");
			}
		},

		/**
		 * First load of the page
		 */
		load: function(name, options) {
			var self = this;
			options = options || {};
			var deferred = $.Deferred();
			var _finalize = function() {
				deferred.resolve();
			};
			var _fail = function(e) {
				self._displayError(e);
				// Deferred is not rejected because error is displayed in the vignette
				_finalize();
			};
			try {
				var gadget = self.prototype.getVignetteGadget();
				if (gadget && gadget.isValid()) {
					switch (gadget.data("$type")) {
						case "$representation":
						case "$stats":
						case "$request":
							self._loadRepresentation(gadget, name, options).then(function() {
								_finalize();
							}).fail(function(e) {
								_fail(e);
							});
							break;
						case "$dashboard":
							self._loadDashboard(gadget);
							break;
						case "$application":
							self._loadApplication(gadget);
							break;
						case "$external":
							self._loadExternal(gadget);
							break;
						case "$gadgetMissing":
							self._displayError("Referenced gadget does not exist: " + gadget.data("$uuid"));
							break;
						default:
							self._displayError("Unsupported gadget type: " + gadget.data("$type"));
							break;
					}
				} else {
					self._displayError("Invalid gadget" + (gadget ? " [" + gadget.getInvalidReason() + "]" : ""), gadget ? gadget.getJSON() : "Null gadget");
				}
				_finalize();
			} catch (e) {
				_fail(e);
			} finally {
				return deferred.promise();
			}
		},

		_loadDashboard: function(gadget) {
			this._addPicker({
				"data-nav": globals.getDashboardName(gadget.data("dashboardName")),
				"data-gadget-id": gadget.data("$uuid"),
				"title": gadget.data("$title")
			});
		},

		_loadApplication: function(gadget) {
			this._addPicker({
				"data-nav": gadget.data("applicationName") + ".$application",
				"data-application-name": gadget.data("applicationName"),
				"title": gadget.data("$title")
			});
		},

		_loadExternal: function(gadget) {
			var self = this,
				url = gadget.data("externalUrl");
			self._addPicker({
				"href": url,
				"title": gadget.data("$title")
			});
			if (self.displayStyle !== "$link") {
				var iFrame = _getHtml("iframe", {
					url: url
				});
				$(iFrame).appendTo(self.$$elmt);
			}
		},

		_addPicker: function(context) {
			if (!context.href) context.href = "#";
			$(_getHtml("linkPage", context)).appendTo(this.$$elmt);
		},

		waitStart: function() {
			uiUtils.waitPluginEnable(this._$$wait);
			this._$$waitBckg.show();
		},
		waitStop: function() {
			uiUtils.waitPluginDisable(this._$$wait);
			this._$$waitBckg.hide();
		},

		refreshBtnsBar: function() {
			var self = this;
			var show = false;
			if (self._linkInside()) {
				if (!self.$$icnBack) {
					self.$$footer.prepend(_getHtml("navbtns"));
					self.$$icnBack = self.$$footer.children("[data-action=vignetteBack]");
					self.$$icnHome = self.$$footer.children("[data-action=vignetteHome]");
				}
				var l = self._stateStack.length;
				self.$$icnBack.toggle(l > 1);
				self.$$icnHome.toggle(l > 2);
				show = self._stateStack.length > 1;
			}
			if (self.enableRefresh) {
				if (!self.$$refresh) {
					self.$$footer.prepend(_getHtml("refreshbtn"));
					self.$$refresh = self.$$footer.children("[data-action=vignetteRefresh]");
				}
				show = true;
			}
			self.$$footer.toggle(show);
			self.$$elmt.toggleClass("footer", show);
		},

		_actVignetteHome: function() {
			var self = this;
			if (self._stateStack.length == 0) return;
			var state = self._stateStack[0];
			self._stateStack = []; // to force clearContent
			self._loadPage(state, null, "home");
		},

		_actVignetteBack: function() {
			var self = this;
			if (self._stateStack.length <= 1) return;
			self._stateStack.pop();
			// !!! true for BACK in order to unstack the current page
			self._loadPage(self._stateStack[self._stateStack.length - 1], null, "back");
		},
		_actVignetteError: function() {
			var self = this;
			modal.error("Vignette load error", self._error);
		},
		_actVignetteRefresh: function() {
			this.refreshPage({});
		},
	});

exports.Klass = _Klass;
