// syracuse-tablet/html/js/controls/fields/ctrlImage.js
"use strict";

var Base = require('syracuse-tablet/html/js/controls/fields/ctrlFieldBase').CtrlFieldBase;
var locale = require('syracuse-tablet/html/js/helpers/locale');
var utils = require('syracuse-tablet/html/js/helpers/utils');
var uiUtils = require('syracuse-tablet/html/js/ui/uiUtils');
var fontUtils = require('syracuse-tablet/html/js/ui/fontUtils');
var native = require('syracuse-tablet/html/js/helpers/native/native');

var _IMG_TYPE = "$SelectedImageType";
/**
 * 
 * 
 * Rendering:
 * 	display mode : ok
 *  edit mode : ok
 * 
 * action :
 * 	change : ok
 *  remove : ok
 *  camera
 *  sign
 *  edit
 *  
 * MetaData:
 * $isReadOnly:  ok
 * $isDisabled:  ok
 * $isHidden: ok
 * 
 */
var _scaleMap = {
	small: {
		width: 640,
		height: 480
	},
	medium: {
		width: 1024,
		height: 768
	},
	large: {
		width: 1920,
		height: 1440
	}
};
var _templates = {
	layout: ' \
		<div class="s-m-img-layout">\
			<section>\
				<a class="img-responsive s-m-field-img" data-nevent-ctrl-click="" href="#" draggable="false">\
				</a>\
			</section>\
			{{#if actions}}\
			<footer> \
				{{#each actions}}\
					{{#if active}}\
						<a class="s-m-meta" \
							href = "#"\
							draggable="false"\
								data-nevent=""\
								data-naction="{{id}}"\
								{{#if display}}\
									style="display:{{display}}">\
								{{/if}}\
							<span \
							{{#if css}}\
								class="{{css}}"\
							{{/if}}\
							>\
							{{#if text}}\
								{{text}}\
							{{/if}}\
							</span>\
						</a>\
					{{/if}}\
				{{/each}}\
			</footer>\
			{{/if}}\
		</div>',
	layoutDispay: '\
			<a class="img-responsive s-m-field-img" href="#" draggable="false">\
			</a>',
	image: '\
		{{#if empty}}\
			<span class="{{css}}"></span>\
		{{else}}\
			<img class="img-responsive" src="{{url}}" draggable=false>\
		{{/if}}\
		',
	inputFile: ' \
		<input\
			type="file"\
			accept="image/*"\
			class="ctrl-evt-change"\
			data-nevent-ctrl-change=""\
		>'
};
var _getHtml = function(name, ctx) {
	var key = name + "compiled";
	var tmpl = _templates[key];
	if (!tmpl) {
		tmpl = _templates[key] = Handlebars.compile(_templates[name]);
	}
	if (!tmpl) alert("template " + name + " not found");
	return tmpl(ctx);
};

exports.CtrlImage = utils.defineClass(

	function CtrlImage(controller, article, prototype, options) {
		options = options || {};
		options.nativeCapabilities = {
			"penAnnotatedImage": true,
			"penSignatureImage": true,
			"camera": true
		};
		Base.call(this, controller, article, prototype, options);
	}, Base, {
		/**
		 * 
		 */
		buildHtml: function(classes) {
			classes = classes || [];
			classes.push("s-m-img");
			Base.prototype.buildHtml.call(this, classes);
		},
		buildFieldValue: function() {
			this._ensure$$value();
			this._clear$$value();
			var imgUrl = this.getValue();
			var empty = imgUrl == null || imgUrl.length === 0;
			if (this.$isEditMode) {
				this.buildEditValue(imgUrl, empty);
			} else {
				this.buildDisplayValue(imgUrl, empty);
			}
			var imagetype = this.controller.dataset.getProperty(this.$bind, _IMG_TYPE);
			if (!empty && imagetype) {
				// Image data us base64 encoded - Comes when page is loaded from draft or template
				this._setImage(imagetype + this.controller.dataset.getJsonValue(this.$bind));
			}
		},
		/**
		 * 
		 */
		buildDisplayValue: function(url) {
			var layout = $(_getHtml("layoutDispay", {})).appendTo(this.$$value);

			this._newImg(layout, url);
		},

		buildEditValue: function(url) {
			var layout = $(_getHtml("layout", {
				actions: this._imageActions(url === null)
			})).appendTo(this.$$value);

			this.img = this._newImg($("a.img-responsive", layout), url);

			$(_getHtml("inputFile", {})).appendTo(this.$$value);
		},
		setDataTypeValue: function(value, type) {
			Base.prototype.setValue.call(this, value);
			if (type) {
				this.controller.dataset.setProperty(this.$bind, _IMG_TYPE, type);
			}
		},
		_imageActions: function(empty) {
			var actions = [{
				id: "$select",
				active: this._isBtnActive("$select"),
				css: "s-m-select",
				text: locale.text("image.label.selectimage")
			}, {
				id: "$edit",
				active: this._isBtnActive("$edit"),
				css: "s-m-edit " + fontUtils.getIconByName("$image-ink")
			}, {
				id: "$sign",
				active: this._isBtnActive("$sign"),
				css: "s-m-sign " + fontUtils.getIconByName("$image-sign")
			}, {
				id: "$camera",
				active: this._isBtnActive("$camera"),
				css: "s-m-camera " + fontUtils.getIconByName("$image-camera")
			}, {
				id: "$delete",
				active: this._isBtnActive("$delete"),
				css: "s-m-delete " + fontUtils.getIconByName("$delete"),
				display: empty ? "none" : null
			}];
			return actions;
		},
		_isBtnActive: function(btnId) {
			var isSignature = this.article.$isSignature ? true : false;
			var isActive = false;
			switch (btnId) {
				case "$select":
					isActive = (!isSignature);
					break;
				case "$edit":
					isActive = (!isSignature && this.isNativeCapabilityEnabled("penAnnotatedImage"));
					break;
				case "$camera":
					isActive = (!isSignature && this.isNativeCapabilityEnabled("camera"));
					break;
				case "$sign":
					isActive = (isSignature && this.isNativeCapabilityEnabled("penSignatureImage"));
					break;
				case "$delete":
					isActive = true;
					break;
				default:
					isActive = false;
			}
			return isActive;
		},
		applyMetaData: function(metaData) {
			Base.prototype.applyMetaData.call(this, metaData);
		},
		onDataActionClicked: function(dataAction, dataParams, event) {
			switch (dataAction) {
				case "$select":
					this._onUploadClick();
					break;
				case "$delete":
					this._onClearClick();
					break;
				case "$edit":
					this._onEditClick();
					break;
				case "$camera":
					this._onCameraClick();
					break;
				case "$sign":
					this._onSignClick();
					break;
				default:
					return false
			}
		},
		onClick: function(event) {
			// Not used - click on image
			// console.log("click on " + $(event.target).prop("nodeName"))
		},
		/*
		 When upload button is clicked (or empty image), run upload and update dao
		 */
		_onUploadClick: function() {
			var $$inputDom = this.$$elmt.find(".s-m-value > input");
			try {
				$$inputDom.trigger("click");
			} catch (e) {
				//In case of wrapper and close the dialog box without any file selected, jquery catch 
				//on statement "Prevent re-triggering of the same event" (jQuery 2.1.1 line 4357) 
				//and does not reinit the property "triggered".
				// So we do it in place of jquery !
				jQuery.event.triggered = undefined;
			}
		},
		/*
		 When clear button is clicked
		 - Removes  content
		 - Hide clear button
		 - Update upload button text if image
		 */
		_onClearClick: function() {
			this._setImage(null);
			this.setValue("");
			var element = this.$$elmt.find(".s-m-value > input");
			element.replaceWith(element = element.clone(true)); // element.val('') does not work with IE
			var clearBtn = this.$$elmt.find(".s-m-img-layout > footer > a[data-naction='$delete']")[0];
			$(clearBtn).toggle();
			if (this._isBtnActive('$edit')) {
				var editBtn = this.$$elmt.find(".s-m-img-layout > footer > a[data-naction='$edit']")[0];
				$(editBtn).toggle();
			}
		},
		/*
		 fire when user select a local image
		 */
		onChange: function(evt) {
			var self = this;
			if (evt.target.nodeName.toLowerCase() === 'input') {
				var files = evt.target.files; // FileList object
				var res = {};
				if (files && files.length) {
					var f = files[0]; // only one file
					if (!f.type.match('image.*')) return;
					res.$contentType = f.type;
					res.$fileName = f.name;
					res.$type = "image";

					var reader = new FileReader();
					/*
					 *  theFile : {
					    	lastModified:1465574128588,
							name:"Tiles.png",
							size:35374,
							type:"image/png"
						}
						content : {
							$contentType:"image/png",
							$fileName:"Tiles.png",
							$type:"image",
							$value:"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAP.." // result from _getImageDataValue
						}
						src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAP.."
						*/
					reader.onload = (function(theFile, content) {
						return function(e) {
							self._checkScale(e.target.result).then(function(src) {
								var dataType = self._getImageDataTypeValue(src);
								if (dataType == null) {
									return;
								}
								//	content.$value = dataType.data;
								// clean "data:image/jpeg;base64,"
								//	delete content.$url; // don't understand, not sure it's necessay
								self.setDataTypeValue(dataType.data, dataType.type); // set the dataset, maybe can avoid it, change is manage by the input type file
								self.setFocus();
								self._setImage(src); // draw the new image
							});
						};
					})(f, res);
					reader.readAsDataURL(f); // read the file and trigger onload at the end
				}
			}
		},
		/*
		 *  src: data:image/png;base64,iVBORw0KGgoAAAAN
		 */
		_checkScale: function(src) {
			if (!src) return $.smResolve(src);
			var self = this;

			var scaler = native.getModule("imageScale");
			if (!scaler) return $.smResolve(src);

			var opts = self._getResizeOptions();
			if (!opts) return $.smResolve(src);
			opts.imageData = src;
			return scaler.getScaledImage(opts)
				.then(function(result) {
					// Change so return  new image
					if (result.action === "ok") {
						return result.imageData;
					}
					// No change, maybe because size was ok already
					if (result.action === "nochange") {
						return src;
					}
					// Here we got an error but we still can return the original image
					return src;
				});
		},
		_getResizeOptions: function() {
			// FD - default is medium otherwise iphone6 picture is too big
			var scale = _scaleMap[this.getAuthoring("$imgScale") || "medium"]
			if (scale == null) {
				return null;
			}
			return {
				width: scale.width,
				height: scale.height,
				mode: "LIMIT_MAX_WIDTH_AND_HEIGHT" // Only supported mode right now
			}
		},
		_getImageDataValue: function(content) {
			var i = content.indexOf("base64,");
			if (i > 0) {
				return content.substring(i + 7);
			} else {
				// don't know when we are in this case ????
				return content.result;
			}
		},
		_getImageDataTypeValue: function(data) {
			if (!data) {
				return null;
			}
			var i = data.indexOf("base64,");
			if (i > 0) {
				return {
					data: data.substring(i + 7),
					type: data.substr(0, i + 7)
				}
			} else {
				// don't know when we are in this case ????
				return null;
			}
		},
		/*
		 Set image based on client image src
		 */
		_setImage: function(src) {
			var core = this.$$elmt.find(".s-m-img-layout > section > a")[0];
			// remove image
			uiUtils.empty(core);
			// update image
			this.img = this._newImg(core, src);
			// toggle clearBtn
			this.$$elmt.find(".s-m-img-layout > footer > a[data-naction='$delete']").toggle(true);
			if (this._isBtnActive('$edit')) {
				this.$$elmt.find(".s-m-img-layout > footer > a[data-naction='$edit']").toggle(true);
			}
		},
		_newImg: function(parent, src) {
			var empty = (src === null);
			var css = empty ? this._getDefaultIcon() + " s-m-empty img-responsive" : "s-m-image";
			var img = $(_getHtml("image", {
				empty: empty,
				css: css,
				url: src,
			})).appendTo(parent);
			return this._setImgHeight(img);
		},
		_defaultImg: function(parent, css) {
			return this._setImgHeight(uiUtils.createDomElement("span", this._getDefaultIcon() + (" " + (css || "")), null, null, parent));
		},
		_setImgHeight: function($i) {
			return $i.height(uiUtils.imgHeight2px(this.getAuthoring("$imgHeight") || "small")).get(0);
		},
		_getDefaultIcon: function() {
			var css = "s-m-icon-" + (this.article.$emptyIconSize || "small") + " ";
			return css + (this.article.$emptyImageIcon || fontUtils.getEmptyImageDefIcon()).css;
		},
		_getBinaryContent: function() {
			var binary;
			if (this.img) {
				var canvas = document.createElement('canvas');
				var ctx = canvas.getContext("2d");
				ctx.canvas.width = this.img.naturalWidth;
				ctx.canvas.height = this.img.naturalHeight;
				ctx.drawImage(this.img, 0, 0, this.img.naturalWidth, this.img.naturalHeight);
				binary = canvas.toDataURL();
			}
			return binary;
		},
		_onEditClick: function() {
			var self = this;
			var editBtn = this.$$elmt.find(".s-m-img-layout > footer > a[data-naction='$edit']")[0];
			var data;

			// Avoid error on invalid image data
			try {
				data = self._getBinaryContent();
			} catch (e) {
				$(editBtn).hide();
				return;
			}
			$(editBtn).toggle();
			native.getModule("penAnnotatedImage").getAnnotatedImage({
				title: self.getTitle(),
				imageData: data
			}).then(
				function(data) {
					if (data && data.action === "ok") {
						if (data.imageData && data.imageData.length) {
							// No need to scale here
							// The image is scaled before adding it and making inking available
							self.setValue(self._getImageDataValue(data.imageData));
							self._setImage(data.imageData);
						}
					} else {
						$(editBtn).toggle();
					};
				}
			);
		},
		_onSignClick: function() {
			var self = this;
			var signBtn = this.$$elmt.find(".s-m-img-layout > footer > a[data-naction='$sign']")[0];
			$(signBtn).toggle();
			native.getModule("penAnnotatedImage").getSignatureImage({
				title: self.getTitle(),
				watermark: ""
			}).then(
				function(data) {
					if (data && data.action === "ok") {
						if (data.imageData && data.imageData.length) {
							// We do not scale images that are input as a signature since
							// signature images created by the wrapper have a defined size already 
							self.setValue(self._getImageDataValue(data.imageData));
							self._setImage(data.imageData);
						}
					}
					$(signBtn).toggle();
				}
			);
		},
		_onCameraClick: function() {
			var self = this;
			var signBtn = this.$$elmt.find(".s-m-img-layout > footer > a[data-naction='$camera']")[0];
			$(signBtn).toggle();
			native.getModule("camera").getPictureFromCamera({
					resizeOptions: self._getResizeOptions()
				}).then(function(data) {
					if (data && data.action === "ok") {
						if (data.imageData && data.imageData.length > 0) {
							return self._checkScale(data.imageData);
						}
					}
					return null;
				})
				.then(function(imageData) {
					if (imageData) {
						self.setValue(self._getImageDataValue(imageData));
						self._setImage(imageData);
					}
				})
				.always(function() {
					$(signBtn).toggle();
				});
		}
	});
