// syracuse-tablet/html/js/controls/fields/ctrlText.js
"use strict";

var Base = require('syracuse-tablet/html/js/controls/fields/ctrlFieldBase').CtrlFieldBase;

var utils = require('syracuse-tablet/html/js/helpers/utils');
var uiUtils = require('syracuse-tablet/html/js/ui/uiUtils');
var fontUtils = require('syracuse-tablet/html/js/ui/fontUtils');

var sdataRequester = require('syracuse-tablet/html/js/sdata/sdataRequester');
var escaper = require('syracuse-tablet/html/js/helpers/html-escape/index').escaper({
	warn: console.warn.bind(console),
	allowedAttributes: ["style"],
	allowedTags: ["span"]
});
var rtf = require('syracuse-tablet/html/js/helpers/syracuse-rtf/index');

var _baseHeight = 34; // based on input height within the app

exports.CtrlText = utils.defineClass(
	function CtrlText(controller, article, prototype, options) {
		options = options || {};
		options.nativeCapabilities = null;
		Base.call(this, controller, article, prototype, options);
		this.cssType = "s-m-clob s-m-" + this.$type.replace("/", "-");
	}, Base, {
		destroy: function() {
			$(this.$$input).off('keyup input');
			Base.prototype.destroy.call(this);
		},
		_getHeightPixelValue: function(authH) {
			switch (authH) {
				case "medium":
					return 2 * _baseHeight;
				case "large":
					return 4 * _baseHeight;
				case "xlarge":
					return 8 * _baseHeight;
				default:
					return _baseHeight;
			}
		},
		buildFieldValue: function() {
			var self = this;
			self._ensure$$value();
			self._clear$$value();
			// dom element which will contain the text
			// enriched text container div
			var $$container = $('<section><div class="s-m-enriched-text"></div></section>').appendTo(self.$$value).children(":first-child");

			// dom element which will contain the text
			var $$textContent = $(uiUtils.createDomElement("div", ["s-m-enriched-text-content"], null, null, $$container));

			/**
			 * !!! It's mandatory to have a fixed height for $$container because we need it to calculate the height of the page
			 * If no height is specified it will be set after the data has been loaded (too late)
			 */
			self.authH = self.getAuthoring("$textHeight", "medium");
			self.authH = self._getHeightPixelValue(self.authH);

			var $url = self.controller.getLazyUrl(self.$bind);
			sdataRequester.getLazyPropData($url, self.prototype.json, self.$type).
			then(function(data) {
				var content = data;
				self.setValue(content);
				var empty = (content == null || content.length === 0) && !self.$isEditMode;
				self.$$elmt.toggleClass("s-m-empty", empty);
				// put text content
				var $$dataArr = [];
				switch (self.$type) {
					case "text/plain":
						self._buildFieldFromPlain(content, $$textContent);
						break;
					case "text/html":
						self._buildFieldFromHtml(content, $$textContent);
						break;
					case "text/rtf":
						self._buildFieldFromRtf(content, $$textContent);
						break;
				}
				/*
				 *	Set text-content height so that line of text is not half cut.
				 *	We use text line-height css property value
				 */
				var lineheight = parseInt($$textContent.css("line-height").replace("px", ""), 10);
				var multiplier = Math.floor(self.authH / lineheight);
				self.computedHeight = lineheight * multiplier;

				$$textContent.css({
					height: self.computedHeight + "px"
				});
				self._appendExpander($$textContent);
			}).fail(function(e) {
				self._insertError(e.message);
			});;
		},

		_buildFieldFromPlain: function(data, $$parent) {
			var self = this;
			if (this.$isEditMode) {
				self.createMainInput(data, "textplain", {
						"wrap": "soft",
						"spellcheck": false,
						"autocorrect": "off",
						"autocomplete": "off",
						"style": "overflow-y:hidden;box-size:border-box;resize:none;width:100%;height:100%",
						"row": "2"
					},
					$$parent);
				this.$$input.val(data);
				this.$$input.on('keyup input', function() {
					self._triggerResizeInput();
				});
			} else {
				$$parent.append($.parseHTML(escaper(data.replace(/\n/g, "<BR/>"))));
			}
		},
		_buildFieldFromHtml: function(data, $$parent) {
			$$parent.append($.parseHTML(escaper(data)));
		},
		_buildFieldFromRtf: function(data, $$parent) {
			this._buildFieldFromHtml(rtf.toHtml(data), $$parent);
		},
		_triggerResizeInput: function() {
			var triggerResize = false;
			if (this.$$expander && this.$$expander.attr("data-params") === "more") {
				this._expandText(this.$$input.parent());
				triggerResize = true;
			};
			this._resizeInput();
			if (triggerResize) {
				uiUtils.triggerResizeInternal({
					preserveScroll: true
				});
			}
		},
		_resizeInput: function(options) {
			if (!this.$$input) return false;
			this.$$input.parent().css('height', 'auto');
			this.$$input.css('height', 'auto');
			var domElmt = this.$$input.get(0);
			var height = (options && options.height != null) ? options.height : (domElmt.scrollHeight + (domElmt.offsetHeight - domElmt.clientHeight));
			this.$$input.css('height', height);
			return false;
		},
		_expandText: function($textContent) {
			$textContent.css({
				height: "auto"
			});
			this.$$expander.toggleClass(fontUtils.getIconByName("expand"), false);
			this.$$expander.toggleClass(fontUtils.getIconByName("compress"), true);
			this.$$expander.attr("data-params", "less");
		},
		_compressText: function($textContent) {
			$textContent.css({
				height: this.computedHeight + "px"
			});
			this.$$expander.toggleClass(fontUtils.getIconByName("compress"), false);
			this.$$expander.toggleClass(fontUtils.getIconByName("expand"), true);
			this.$$expander.attr("data-params", "more");
		},
		/*
		 * Toggle enriched text. It will either show more or hide
		 */
		_actToggleText: function(param) {
			// show whole text field
			// change button label
			var $textContent = this.$$input ? this.$$input.parent() : this.$$elmt.find(".s-m-enriched-text-content");
			if (param === "more") {
				this.currentHeight = null;
				this._expandText($textContent);
				this._resizeInput();
			}
			// decrease text field height
			else {
				this._compressText($textContent);
				this._resizeInput({
					"height": "100%"
				});
			}
			uiUtils.triggerResizeInternal({
				preserveScroll: true
			});
		},


		/**
		 * Display an error
		 */
		_insertError: function(msg) {
			var err = ['<span class="s-m-error"><p>'];
			err.push(msg);
			err.push("<p></span>");
			$(err.join('')).appendTo(this.$$elmt);
		},
		/*
		 *	Append an expander button for the text field.
		 *	Button is appended only if full text height (scrollheight) is higher than self.authH by at least 5px
		 */
		_appendExpander: function($$container) {
			var self = this;
			var scrollHeight = self.$$input ? self.$$input.scrollHeight : $$container.get(0).scrollHeight;
			if (scrollHeight > self.computedHeight && ((scrollHeight - self.computedHeight) > 5)) {

				// build expander button
				var buttonTmpl = '<a href="#" class="' + fontUtils.getIconByName("expand") + '" data-nevent data-naction="toggleText"\
				data-control-id="' + self.id + '" data-params="more"></a>';
				var $$expander = self.$$expander = $(buttonTmpl);

				// append the button
				$$expander.insertBefore($$container);

				$$expander.css("float", "right");

				$$expander.wrap("<div class='s-m-enriched-text-expander'></div>");

				$$expander.parent().height($$expander.outerHeight());
			}
		}
	});
