// @sage/syracuse-lib/src/collaboration/entities/user/oauth2.ts
"use strict";

const locale = require('streamline-locale');
import * as util from 'util';
import { AdminHelper as adminHelper } from '../../../collaboration/helpers';
import * as config from 'config';

export const entity = {
	$properties: {
		name: {
			$title: "Name",
			$isMandatory: true,
			$default: "OAuth2",
			$linksToDetails: true,
			$isUnique: true,
			$pattern: "^[a-zA-Z]\\w*$",
		},
		displayName: {
			$title: "Display Name",
			$isLocalized: true
		},
		active: {
			$title: "Active",
			$type: "boolean",
			$default: true
		},
		baseSite: {
			$title: "OAuth2 server URL without path",
			$isMandatory: true
		},
		authorizePath: {
			$title: "Path for authorization",
			$default: "/oauth/authorize",
			$isMandatory: true
		},
		accessTokenPath: {
			$title: "Path to get access token",
			$isMandatory: true,
			$default: "/oauth/access_token"
		},
		clientId: {
			$title: "OAuth2 client ID",
			$isMandatory: true
		},
		clientSecret: {
			$title: "OAuth2 client secret",
			$type: "password",
			$encrypt: true,
			$isMandatory: true
		},
		scope: {
			$title: "Scope for OAuth2 requests"
		},
		batchAuthentication: {
			$title: "Batch authentication",
			$type: "boolean",
			$default: false
		},
		redirectPath: {
			$title: "Redirect path for OAuth2 server",
			$compute(_, instance) {
				return "/auth/oauth2/" + instance.name(_) + "/loginCallback";
			},
			$isHidden(_) {
				return (config.hosting && config.hosting.multiTenant) ? true : false;
			}
		},
		apiPath: {
			$title: "Web API route path for OAuth2 redirect",
			$compute(_, instance) {
				return "/api/" + instance.name(_) + "/oauth2/redirect";
			},
			$isHidden(_) {
				return (config.hosting && config.hosting.multiTenant) ? false : true;
			}
		},
		dataRequestURL: {
			$title: "URL for requesting user data",
			$isMandatory: true
		},
		userField: {
			$title: "User field in user name answer",
			$isMandatory: true,
			$default: "user"
		},
		image: {
			$title: "Image",
			$type: "image",
			$storage: "db_file",
			$capabilities: ""
		},
		jwtIssuer: {
			$title: "JWT Issuer",
			$compute(_, instance) {
				return instance.clientId(_);
			}
		},
		jwtAudience: {
			$title: "JWT Audience",
			$compute(_, instance) {
				return instance.baseSite(_);
			}
		}
	},
	$titleTemplate: "OAuth2",
	$valueTemplate: "{name} {baseSite}",
	$descriptionTemplate: "OAuth2 server {name}",
	$helpPage: "Administration-reference_OAuth2",
	$relations: {
		users: {
			$title: "Users",
			$type: "users",
			$inv: "oauth2",
			$isComputed: true
		},
		certificate: {
			$title: "JWT Certificate",
			$type: "certificate",
			$inv: "oauth2"
		}
	},
	$searchIndex: {
		$fields: ["name", "displayName", "baseSite"]
	},
	$defaultOrder: [
		["name", true]
	]
};
