// syracuse-ui/lib/help/helpCenter.js
"use strict";
var LegalPage = require('syracuse-ui/lib/help/legalPage').LegalPage;
var List = require('syracuse-ui/lib/scroll/list').List;
var _list, _popup;
var _browserTab = {};

exports.openByShortCut = function(page, field) {
	page.externalAdapter.onFieldClickPicker({
		field: field || page,
		pickerType: "help",
		doEvent: function() {
			exports.open(page, field);
		}
	});
};

exports.open = function(page, field, $url) {
	var winId = "sageerphelp";
	if (!$url) {
		var $links = page.$prototype.$links;
		$url = $links && $links.$help && $links.$help.$url;
		if ($url) {
			if (field) {
				$url = syra_expression.parse(field.articleParent, $url) + '#' + field.$item.$bind;
			} else {
				$url = syra_expression.parse(page, $url);
			}
		}
	}
	if ($url) {
		if (syra_context.browser.isMSIE) {
			// Troubles with IE... to improve!
			window.open($url, "_blank");
			return;
		}
		if (!_browserTab.winRef || _browserTab.winRef.closed == undefined || _browserTab.winRef.closed) {
			_browserTab.winRef = window.open($url, winId, "width=800,height=600,resizable,scrollbars,status");
		} else
		if (_browserTab.lastUrl != $url) {
			_browserTab.winRef = window.open($url, winId, "resizable=true,scrollbars=yes,status=yes");
			if (_browserTab.winRef && _browserTab.winRef.focus) {
				_browserTab.winRef.focus();
			}
		} else
		if (_browserTab.winRef && _browserTab.winRef.focus) {
			_browserTab.winRef.focus();
		}
		_browserTab.lastUrl = $url;
	}
};

exports.openPopup = function(menuItem) {
	if (!_list) {
		_list = new List({
			css: "s-list-primary"
		});
		_list.addItem(syra_site, {
			$bind: "site_help_center",
			$category: "link",
			$skin: "s-list-primary-btn-default"
		});
		_list.addItem(syra_site, {
			$bind: "site_help_shortcuts",
			$category: "link",
			$skin: "s-list-primary-btn-default"
		});
	}
	if (!_popup) {
		_popup = syra_over.openPopup(syra_site, {
			content: syra_site,
			slot: _list.scrollSlot,
			picker: menuItem.domItem,
			position: {
				my: "left top",
				at: "left bottom",
				of: menuItem.domItem
			},
			onresize: function(maxHeight) {
				_list && _list.resize(maxHeight);
			},
			close: function() {
				_popup = null;
			}
		});
	} else {
		_popup.close();
	}
};

exports.openLegalPage = function(menuItem) {
	syra_ajax.get({
		url: syra_config.$legalPrototypeUrl,
		success: function(data, response, $url) {
			var page = new LegalPage();
			page.load(data);
			syra_ajax.get({
				url: syra_config.$legalDataUrl,
				success: function(data, response, $url) {
					page.applyChange(data);
					syra_over.openModal(menuItem.page, {
						page: page
					});
				}
			});

		}
	});

};

exports.dispose = function() {
	_list && _list.dispose();
	_browserTab = _list = _popup = null;
};
