// syracuse-ui/lib/help/helpCenter.js
"use strict";
var LegalPage = require('syracuse-ui/lib/help/legalPage').LegalPage;

var _popupSlot, _popup;
var _browserTab = {};

exports.openByShortCut = function(page, field) {
	page.externalAdapter.onFieldClickPicker({
		field: field || page,
		pickerType: "help",
		doEvent: function() {
			exports.open(page, field);
		}
	});
};

exports.open = function(page, field, $url) {
	var winId = "sageerphelp";
	if (!$url) {
		var $links = page.$prototype.$links;
		$url = $links && $links.$help && $links.$help.$url;
		if ($url) {
			if (field) {
				$url = syra_expression.parse(field.articleParent, $url) + '#' + field.$item.$bind;
			} else {
				$url = syra_expression.parse(page, $url);
			}
		}
	}
	if ($url) {
		if (syra_site.browser.isMSIE) {
			// Troubles with IE... to improve!
			window.open($url, "_blank");
			return;
		}
		if (!_browserTab.winRef || _browserTab.winRef.closed == undefined || _browserTab.winRef.closed) {
			_browserTab.winRef = window.open($url, winId, "width=800,height=600,resizable,scrollbars,status");
		} else
		if (_browserTab.lastUrl != $url) {
			_browserTab.winRef = window.open($url, winId, "resizable=true,scrollbars=yes,status=yes");
			if (_browserTab.winRef && _browserTab.winRef.focus) {
				_browserTab.winRef.focus();
			}
		} else
		if (_browserTab.winRef && _browserTab.winRef.focus) {
			_browserTab.winRef.focus();
		}
		_browserTab.lastUrl = $url;
	}
};

exports.openPopup = function(menuItem) {
	if (!_popupSlot) {
		_popupSlot = syra_dom.addDiv("s-mn-popup-body");
		syra_site.loadNewItem(_popupSlot, {
			$bind: "$documentation",
			$category: "link",
			$skin: "s-mn-link"
		});
		syra_site.loadNewItem(_popupSlot, {
			$bind: "$shortcuts",
			$category: "link",
			$skin: "s-mn-link"
		});
	}
	if (!_popup) {
		_popup = syra_dlg.openPopup(syra_site, {
			content: syra_site,
			slot: _popupSlot,
			picker: menuItem.domItem,
			position: {
				my: "left top",
				at: "left bottom",
				of: $(menuItem.domItem)
			},
			onClose: function() {
				_popup = null;
			}
		});
	} else {
		_popup.close();
	}
};

exports.openLegalPage = function(menuItem) {
	syra_http.get({
		url: syra_site.$legalPrototypeUrl,
		success: function(data, response, $url) {
			var page = new LegalPage();
			page.loadBox(data);
			syra_http.get({
				url: syra_site.$legalDataUrl,
				success: function(data, response, $url) {
					page.applyChange(data);
					syra_dlg.openModal(menuItem.page, {
						page: page
					});
				},
				error: function(error) {
					if (error.data.indexOf("$diagnoses") != -1) {
						syra_alert.error(JSON.parse(error.data).$diagnoses);
					} else {
						syra_alert.error(error.data);
					}
				}
			});

		},
		error: function(error) {
			if (error.data.indexOf("$diagnoses") != -1) {
				syra_alert.error(JSON.parse(error.data).$diagnoses);
			} else {
				syra_alert.error(error.data);
			}
		}
	});

};

exports.dispose = function() {
	_browserTab = _popupSlot = _popup = null;
};
