// syracuse-ui/lib/help/legalPage.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var DesktopPage = require("syracuse-ui/lib/page/desktopPage").DesktopPage;
var _$eulaUrl = "/syracuse-ui/lib/help/SAGE X3 EULA.pdf";

function LegalPage() {}

exports.LegalPage = helpers.defineClass(LegalPage, DesktopPage, {
	loadBox: function($prototype) {
		this.isMenuBarDisabled = true;
		this.isFooterDisabled = true;
		this.layoutSlot = document.createElement("div");
		this.$prototype = $prototype;
		this.$prototype.$properties.$banner = {
			$type: "application/x-string"
		};

		this.$prototype.$links = this.$prototype.$links || {};
		this.$prototype.$links.$eula = {
			$title: syra_local.legalCopyright,
			$description: syra_local.legalEula,
			$target: "blank",
			$url: _$eulaUrl
		};
		this.$prototype.$links.$banner = {
			$isDisabled: true,
			$title: "Legal banner"
		};

		var $skin = "s-site-legal";

		this.$item = {
			$category: "page",
			$css: $skin + "-header",
			$title: syra_local.legalPageTitle,
			$layout: {
				$items: [{
					$format: "cards",
					$clientId: "s-legal-page-list",
					$addMaximize: false,
					$isTitleHidden: true,
					$isPagerHidden: true,
					$isQuickDesignerDisabled: true,
					$isRowIndexVisible: false,
					$selectByRowIndex: false,
					$isMenuRecordHidden: true,
					$bind: "$resources",
					$isHoverDisabled: true,
					$skin: $skin + "-card",
					renderRecordContent: function(record) {
						var head = syra_dom.addDiv("s-legal-item-head");
						syra_dom.addTextInfo("s-legal-item-product", record.dataset.product, head);
						syra_dom.addTextInfo("s-legal-item-version", record.dataset.productVersion, head);
						record.body.appendChild(head);
						var $rows = [
							["serial", "valEnd"],
							["licenseeName", "resellerName"],
							["licenseeRegistrationNumber", "resellerRegistrationNumber"]
						];
						var tb = document.createElement("div");
						tb.className = "s-legal-item-body";
						for (var ii = 0, jj = $rows.length; ii < jj; ii++) {
							var $cells = $rows[ii];
							var row = document.createElement("div");
							row.className = "s-legal-item-row";
							for (var mm = 0, kk = $cells.length; mm < kk; mm++) {
								if (mm > 0) {
									syra_dom.addDiv("s-legal-item-field-sep", row);
								}
								var $bind = $cells[mm];
								syra_dom.addTextInfo("s-legal-item-field-title", record.$prototype.$properties[$bind].$title, row);
								record.page.loadNewItem(syra_dom.addDiv("s-legal-item-field-value", row), {
									$bind: $bind,
									$inplace: true
								}, record);

							}
							tb.appendChild(row);
						}
						record.body.appendChild(tb);
					},
					$layout: {
						$items: []
					}

				}]
			}
		};
		DesktopPage.prototype.loadBox.call(this);
		var list = this.idMap["s-legal-page-list"];
		syra_dom.hide(list.topbar, true);

		this.loadNewItem(syra_dom.addDiv("s-legal-footer", this.layoutContent.domItem), {
			$category: "link",
			$skin: "s-site-legal-copyright",
			$bind: "$eula"
		});
	},
	applyChange: function(newData) {
		if (newData) {
			if (newData.$resources && newData.$resources.length > 0) {
				for (var ii = 0, jj = newData.$resources.length; ii < jj; ii++) {
					newData.$resources[ii].productVersion = syra_local.legalProductUpdateLabel + " " + newData.$resources[ii].productVersion;
				}
			}
			if (this.dataset && this.dataset.$copyright) {
				newData.$links = newData.$links || {};
				(newData.$links.$eula = newData.$links.$eula || {}).$title = this.dataset.$copyright;
			}
			DesktopPage.prototype.applyChange.call(this, newData);
		}
	},
	onMenuClick: function(menuItem) {
		if (menuItem.$bind == "$eula") {
			window.open(_$eulaUrl, "_blank");
			return false;
		}
		return true;
	}
});
