// syracuse-ui/lib/office/ppt/pptSite.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Site = require('syracuse-ui/lib/site/site').Site;
var Controller = require("syracuse-ui/lib/controller/controller").Controller;
var store = require('syracuse-ui/lib/site/tools/store');

function PptSite() {}

exports.PptSite = helpers.defineClass(PptSite, Site, {
	_onAfterLogon: function() {
		syra_site.history.start();
		try {
			this.pptInterface.onLogin();
		} catch (e) {}
		var addinVersion = this.pptInterface.getAddinVersion();
		this.layoutSlot.style.display = "";
		this.resize();
	},
	_renderHeader: function() {

	}
});

exports.load = function($item, $prototype) {
	window.syra_store = store;
	var widgetsLibrary = require('syracuse-ui/lib/site/widgetsLibrary');
	widgetsLibrary.pageCategories.ppt_populate_worksheet = require("./pptPage").PptPage;
	$item.$isSecurityTitleVisible = false;
	var site = (new PptSite()).loadBox({
		isOffice: true,
		isUserEndpointChangeDisabled: true,
		widgetsLibrary: widgetsLibrary,
		controllerClass: Controller,
		$item: $item,
		$prototype: $prototype
	});

	var pptInterface = require('msoffice/lib/ppt/pptInterface').PptInterface;
	site.pptInterface = new pptInterface();
	site.$syracuseMainPageUrl = $item.$syracuseMainPageUrl;
	return site;
};

//Called from PowerPoint to close the connection to Syracuse
//ONLY WORKS IN IE. no problem here, since this is only called from IE within word!
window.postUrl = function(url, returnData) {
	try {
		var xhr = new XMLHttpRequest();
		var data = {};
		xhr.open('POST', url, false);

		xhr.setRequestHeader("X-Content-Type-Override", "text/plain; charset=x-user-defined");
		xhr.onreadystatechange = function(e) {};
		xhr.send();
		if (xhr.status == 200) {
			// this.responseText is broken in IE with binary data!
			// This is a c# byte array!
			returnData.setData(xhr.responseBody);
		} else
		if (xhr.status == 1223) {
			returnData.setErrorText("NOTFOUND");
		} else
		if (xhr.status == 404) {
			returnData.setErrorText("NOTFOUND");
		} else {
			returnData.setErrorText(xhr.response + "(" + xhr.status + ")");
		}
	} catch (e) {
		returnData.setErrorText(e);
	}
};
