// syracuse-ui/lib/office/ppt/pptSite.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Site = require('syracuse-ui/lib/site/site').Site;
var Controller = require("syracuse-ui/lib/controller/controller").Controller;

function PptSite(){
}

exports.PptSite = helpers.defineClass(PptSite, Site, {
    _onLogon: function(){
        try {
            this.pptInterface.onLogin();
        } 
        catch (e) {
        }
        var addinVersion = this.pptInterface.getAddinVersion();
    },
    updateDocumentTitle: function(){
        var title = this.mainPage.getTitle();
        if (title) {
            var record = this.userProfile.dataset;
            this.mainPage.drawUserProfile(record);
            
            document.title = title;
        }
    },
    _renderHeader: function(){
        this.$$userProfileHost = $("<div id='s-user-profile'/>").appendTo(this.$$container);
    },
    _renderSearch: function($$container){
    }
});

exports.load = function($item, $prototype){
    var self = this;
    
    var widgetsLibrary = require('syracuse-ui/lib/site/widgetsLibrary');
    widgetsLibrary.pageCategories.ppt_populate_worksheet = require("./pptPage").PptPage;
    var site = (new PptSite()).loadBox({
        widgetsLibrary: widgetsLibrary,
        userProfileClass: require("syracuse-ui/lib/office/userProfile").UserProfile,
        controllerClass: Controller,
        $item: $item,
        $prototype: $prototype
    });
    
    var pptInterface = require('msoffice/lib/ppt/pptInterface').PptInterface;
    site.pptInterface = new pptInterface();
    site.$syracuseMainPageUrl = $item.$syracuseMainPageUrl;
    return site;
};
