// syracuse-ui/lib/office/ppt/pptSite.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Site = require('syracuse-ui/lib/site/site').Site;
var Controller = require("syracuse-ui/lib/controller/controller").Controller;

function PptSite(){
}

exports.PptSite = helpers.defineClass(PptSite, Site, {
    drawBox: function(){
        var self = this;
        self._renderHeader();
        self.$$body = $("<div id='s-site-body'/>").appendTo(self.$$container);
        $(window).bind("resize", function(evt){
            self.resize();
        });
        
        document.site.logon(function(){
            document.controller.startNavigation();
            self.$$container.show();
            self.resize();
            
            try {
                self.pptInterface.onLogin();
            } 
            catch (e) {
            }
            
            var addinVersion = self.pptInterface.getAddinVersion();
        });
    },
    
    updateDocumentTitle: function(){
        var title = this.mainPage.getTitle();
        if (title) {
            var record = this.userProfile.dataset;
            this.mainPage.drawUserProfile(record);
            
            document.title = title;
        }
    },
    _renderHeader: function(){
        this.$$userProfileHost = $("<div id='s-user-profile'/>").appendTo(this.$$container);
    },
    _renderSearch: function($$container){
    }
});

exports.load = function($item, $prototype){
    var self = this;
    
    var widgetsLibrary = require('syracuse-ui/lib/site/widgetsLibrary');
    widgetsLibrary.defaultPageCategory = OfficePage;
    widgetsLibrary.pageCategories.ppt_populate_worksheet = require("./pptPage").PptPage;
    var site = (new PptSite()).loadSite({
        widgetsLibrary: widgetsLibrary,
        userProfileClass: require("./userProfile").UserProfile,
        controllerClass: Controller,
        $item: $item,
        $prototype: $prototype
    });
    
    var pptInterface = require('msoffice/lib/ppt/pptInterface').PptInterface;
    site.pptInterface = new pptInterface();
    site.$syracuseMainPageUrl = $item.$syracuseMainPageUrl;
    return site;
};
