// syracuse-ui/lib/office/ppt/pptSaveSite.js
"use strict";

// This site is only used to exchange the controller an catch the save event of a document entry to directly
// save content from an active document inside ms ppt

var helpers = require('syracuse-core/lib/helpers');
var Site = require('syracuse-ui/lib/site/site').Site;
var OfficePage = require('syracuse-ui/lib/office/officePage').OfficePage;
var CommonController = require('syracuse-ui/lib/controller/controller').Controller;

function PptSafeSite(){
}

exports.PptSafeSite = helpers.defineClass(PptSafeSite, Site, {
    drawBox: function(){
        var self = this;
        self._renderHeader();
        self.$$body = $("<div id='s-site-body'/>").appendTo(self.$$container);
        $(window).bind("resize", function(evt){
            self.resize();
        });

        document.site.logon(function() {
            document.controller.startNavigation();
            self.$$container.show();
            self.resize();

            try {
            	// Access the current document url (URL where the doc is saved in storage area)
            	// if the url  is set, directly safe w/o showing edit page
            	var docUrl = document.site.pptInterface.getDocumentUrl();
            	if (docUrl && docUrl !== "")
            		document.site.pptInterface.publishDocument();
            } catch(e) {
            }
        });
    },
    updateDocumentTitle: function(){
        var title = this.mainPage.getTitle();
        if (title) {
            var record = this.userProfile.dataset;
            this.mainPage.drawUserProfile(record);
            
            document.title = title;
        }
    },
    _renderHeader: function(){
        this.$$userProfileHost = $("<div id='s-user-profile'/>").appendTo(this.$$container);
    },
    _renderSearch: function($$container){
    }
});

exports.load = function($item, $prototype) {
	var self = this;
	
    var widgetsLibrary = require('syracuse-ui/lib/site/widgetsLibrary');
    widgetsLibrary.defaultPageCategory = OfficePage;
    var site = (new PptSafeSite()).loadSite({
        widgetsLibrary: widgetsLibrary,
        userProfileClass: require("./userProfile").UserProfile,
        controllerClass: CommonController,
        $item: $item,
        $prototype: $prototype
    });

    var pptInterface = require('msoffice/lib/ppt/pptInterface').PptInterface;
    site.pptInterface = new pptInterface();
    site.$syracuseMainPageUrl = $item.$syracuseMainPageUrl;
    return site;
};
