// syracuse-tablet/html/js/controls/panels/header/panelHeaderEditHome.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var Base = require('syracuse-tablet/html/js/controls/panels/header/panelHeaderSdataPageBase').PanelHeaderSdataPageBase;
var notifications = require('syracuse-tablet/html/js/helpers/notifications');
var jsutils = require('syracuse-tablet/html/js/helpers/jsutils');
var modules = require('syracuse-tablet/html/js/common/modules');
/**
 * Dedicated header for edit homepage on tablet
 */
var _html = {
	mainContainer: '\
		<div class="s-m-header {{device}} s-m-dasboard-config">\
			<buttons class="pull-left dashboard-header-btn"> \
				<button type="button" class="btn btn-default fa fa-arrow-left" data-nevent data-naction="closeDashboardConfig" data-controller-id="{{controllerId}}"></button> \
			</buttons> \
			<div class="s-m-header-title">\
				<input class="s-m-header-label" value="{{description}}">\
				{{#if template}}\
					<div class="s-m-dashboard-header-role-number">{{roleCount}}</div>\
				    <btn class="s-m-dashboard-header-btn-role fa fa-users" data-nevent data-naction="roleList" data-controller-id="{{controllerId}}">\
				{{/if}} \
			</div>\
			<buttons class="pull-right dashboard-header-btn" style="width:{{rightBtnWidth}}px">\
				<button type="button" class="btn btn-default fa fa-check" style="display:{{displaySaveButton}}" data-nevent data-naction="{{buttonSaveAction}}" data-controller-id="{{controllerId}}"></button>\
			</buttons>\
		</div>'
}
var _templates = {};

function _getHtml(name, ctx) {
	var tmpl = _templates[name];
	if (!tmpl) {
		tmpl = _templates[name] = Handlebars.compile(_html[name]);
	}
	return tmpl(ctx);
};
exports.PanelHeaderEditHome = utils.defineClass(
	function PanelHeaderEditHome(controller, type, article, opts) {
		Base.call(this, controller, type, article, opts);
		notifications.subscribe(this, ["sm.dashboard.state.change"]);
		notifications.subscribe(this, ["sm.dashboard.roles.change"]);
	}, Base, {

		destroy: function() {
			Base.prototype.destroy.call(this);
		},
		buildHtml: function() {
			Base.prototype.buildHtml.call(this);
			this.$$elmt.html(_getHtml("mainContainer", this.getHeaderCtx()));
			$("input.s-m-header-label", this.$$elmt).on('change', jsutils.bindFn(this._descriptionChange, this))
		},
		_descriptionChange: function() {
			notifications.publish(["sm.dashboard.description.change"], $("input.s-m-header-label", this.$$elmt).val());
		},
		notifDashboardStateChange: function(dirty) {
			if (dirty) {
				$("[data-naction='savePersonal'],[data-naction='saveTemplate']", this.$$elmt).show();
			} else {
				$("[data-naction='savePersonal'],[data-naction='saveTemplate']", this.$$elmt).hide();
			}
		},
		notifDashboardRolesChange: function() {
			$(".s-m-dashboard-header-role-number", this.$$elmt).text(this.page.roles().length);
		},
		/**
		 * Override _getItems
		 */
		getHeaderCtx: function() {
			return {
				description: this.page.getPageDescription(),
				device: modules.get("siteLayout").getDeviceType(),
				displaySaveButton: "none",
				rightBtnWidth: 50,
				buttonSaveAction: this.page.isPersonal() ? "savePersonal" : "saveTemplate",
				controllerId: this.page.controller.id,
				template: !this.page.isPersonal(),
				roleCount: this.page.roles().length
			};
		},
	});
