// syracuse-tablet/html/js/controls/panels/header/panelHeaderSdataPageBase.js
"use strict";

var Base = require('syracuse-tablet/html/js/controls/panels/panelSdataPage').PanelSdataPage;

var formatApi = require('syracuse-tablet/html/js/helpers/formatApi');
var utils = require('syracuse-tablet/html/js/helpers/utils');
var uiSettings = require('syracuse-tablet/html/js/ui/uiSettings');
var notifications = require('syracuse-tablet/html/js/helpers/notifications');
var modules = require('syracuse-tablet/html/js/common/modules');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var Hammer = require('syracuse-tablet/html/deps/hammerjs-2.0.4/hammer');
var globals = require('syracuse-tablet/html/js/app/globals');
var siteLayout = require('syracuse-tablet/html/js/ui/siteLayout');
/**
 * 
 */
var _html = {
	mainContainer: '\
			<div class="s-m-header {{device}}">\
			{{#if showNav}}\
				<div class="s-m-header-nav">\
				<a href="#" draggable="false"\
					id="{{navigation.id}}"\
					data-nevent\
					data-naction="{{navigation.action}}">\
					<i class="{{navigation.icon}}"></i>\
				</a>\
				</div>\
			{{/if}}\
			{{#if showLabel}}\
				<div class="s-m-header-labels">\
					<span class="s-m-header-label">{{title}}</span>\
					<!--\
					<div class="s-m-header-brand">\
						<span class="s-m-brand">Sage</span>\
						<span class="s-m-product">X3</span>\
					</div>\
					-->\
				</div>\
			{{/if}}\
			<div class="s-m-header-links">\
				{{#if showActionPanel}}\
				<a class="s-m-header-link" href="#" draggable="false" \
					data-nevent\
					data-naction="openActionsPanel">\
					<i class="fa fa-ellipsis-v"></i>\
				</a>\
				{{/if}}\
			</div>\
		</div>\
		{{#if showFreshness}}\
			<div class="s-m-freshness s-m-fresh"\
				data-nevent data-naction="showFreshness"\
				style="display: none;">\
				<i class="fa fa-bookmark"></i>\
			</div>\
		{{/if}}'
}
var _templates = {};

function _getHtml(name, ctx) {
	var tmpl = _templates[name];
	if (!tmpl) {
		tmpl = _templates[name] = Handlebars.compile(_html[name]);
	}
	return tmpl(ctx);
};

exports.PanelHeaderSdataPageBase = utils.defineClass(
	function PanelHeaderSdataPageBase(controller, type, article, opts) {
		Base.call(this, controller, type, article, opts);
		notifications.subscribe(this, ["sm.data.freshness.change"]);
	}, Base, {

		destroy: function() {
			Base.prototype.destroy.call(this);
			if (this.hammer) {
				this.hammer.destroy();
				this.hammer = null;
			}
			notifications.unsubscribe(this, ["sm.data.freshness.change"]);
		},

		///
		/// START: Rendering
		///

		/**
		 * 
		 */
		buildHtml: function() {
			this.$$container.css({
				left: 0 + "px",
				height: this.getDesiredHeight() + "px"
			});

			Base.prototype.buildHtml.call(this);
			this.$$elmt.html(_getHtml("mainContainer", this.getHeaderCtx()));
			this.$$elmt.addClass("header");
			this.$$elmt.css({
				left: 0 + "px",
				height: this._height + "px"
			});
			this.$$container.show();
			this._addPressHandler()
		},
		_addPressHandler: function() {
			if (this.hammer) {
				this.hammer.destroy();
			}
			this.hammer = new Hammer(this.$$elmt.get(0));
			this.hammer.get("press").set({
				time: 400 //Small timeout needed to stop immediately when press
			});
			this.hammer.on("press", Hammer.bindFn(this._onPress, this));
		},
		_onPress: function(evt) {
			if (this.page.isHome()) return
			var panel = this.page.openPanel("breadcrumbs");
			if (panel) panel.show();
		},
		/**
		 * 
		 */
		computeLayout: function(context) {
			if (!this.$$container) {
				return;
			}
			var layoutInfo = modules.get("siteLayout").getLayoutInfo();
			this.$$container.css({
				top: this.top + "px",
				width: layoutInfo.application.width + "px",
			});
		},
		///
		/// END: Rendering
		///

		/*
		 * Use by smartphone and tablet to get default information shown in the header
		 */
		getHeaderCtx: function() {
			return {
				title: this.actionAdapter.getPageTitle(),
				navigation: {
					id: (this.actionAdapter.showGlobalPanelButton ? "s-m-header-panel-id" : "s-m-header-back-id"),
					action: (this.actionAdapter.showGlobalPanelButton ? "openGlobalPanel" : "historyBack"),
					icon: (this.actionAdapter.showGlobalPanelButton ? "fa fa-bars" : "fa fa-arrow-circle-o-left"),
				},
				showNav: true,
				showLabel: true,
				showFreshness: true,
				showActionPanel: (!this.actionAdapter.showGlobalPanelButton),
				device: siteLayout.getDeviceType()
			};
		},
		_updateWidth: function() {
			return;
			var $$header = this.$$elmt.find(".s-m-header");
			if (!$$header) {
				return;
			}
			var $$label = this.$$elmt.find(".s-m-header-labels");
			var $$freshness = this.$$elmt.find(".s-m-freshness");
			var freshVisible = $$freshness && $$freshness.is(":visible");
			if (freshVisible) {
				// We need to hide $$freshness to calculate the width
				// $$freshness is positioned in absolute and is dispayed over the header
				$$freshness.hide();
			}

			var width;
			var maxWidth = this.$$elmt.width();
			var fullWidth = 0;
			var $$nav;
			$$header.children().each(function() {
				$$nav = $(this);
				if (!$$nav.is(".s-m-header-labels")) {
					width = 0;
					$$nav.children().each(function(idx) {
						var c = $(this).outerWidth() + 1;
						width += c;
					});
					// Width=with or remaining with or zero
					width = Math.min(Math.max(maxWidth - fullWidth, 0), width);
					fullWidth += width;
					$$nav.width(width);
				}
			});
			$$label.width("auto");
			if (freshVisible) {
				$$freshness.show();
			}
		},
		getDesiredHeight: function() {
			return uiSettings.getProp("header.height");
		},
		setTop: function(top) {
			this.top = top;
		},
		notifDataFreshnessChange: function(dataFreshness) {
			this._dataFreshness = dataFreshness;
			this._$$freshness = $(".s-m-freshness", this.$elmt);
			this._$$freshness.removeClass("s-m-fresh s-m-tainted s-m-verytainted");
			this._$$freshness.addClass("s-m-" + this._dataFreshness.level);
			this._$$freshness.show();
		},

		/**
		 * Displays freshness info
		 */
		_actShowFreshness: function() {
			if (!this._dataFreshness) return;
			var formatter = formatApi.getFormatter("application/x-datetime");
			var dateTime = formatter.formatValue(utils.getCurISODateTime(new Date(this._dataFreshness.dateTime)), false);

			var notify = {
				severityClass: this._dataFreshness.level,
				title: locale.text("page.data.freshness.last") + " : " + dateTime,
				body: locale.text("page.data.freshness." + this._dataFreshness.level),
				onlyWebapp: true
			};
			modules.get("modal").notify(notify);
		}
	});
