// syracuse-tablet/lib/representations/mobileApplication.js
"use strict";
exports.representation = {
	$entityName: "mobileApplication",
	$facets: {
		$details: {
			$layout: {
				$layoutType: "row",
				$items: [{
					$items: [{
						$category: "section",
						$title: "{@InformationsSectionTitle}",
						$layout: {
							$items: [{
								$bind: "applicationName"
							}, {
								$bind: "title"
							}, {
								$bind: "description"
							}, {
								$bind: "$factory"
							}, {
								$bind: "$factoryOwner"
							}]
						}
					}, {
						$category: "section",
						$title: "{@ContentSectionTitle}",
						$layout: {
							$items: [{
								$bind: "iconName",
							}, {
								$bind: "onlineOnly",
							}, {
								$bind: "useHomeGadget",
							}, {
								$bind: "homeDashboard",
							}, {
								$bind: "homeGadget",
							}]
						}
					}, {
						$category: "section",
						$title: "{@AdministrationSectionTitle}",
						$layout: {
							$items: [{
								$bind: "roles",
							}, {
								$bind: "endPoints",
							}, {
								$bind: "representations",
							}]
						}
					}]
				}]

			}
		},
		$edit: {
			$copy: "$details"
		},
		$query: {
			$layout: {
				$items: [{
					$category: "section",
					$layout: {
						$items: [{
							$category: "section",
							$layout: {
								$items: [{
									$bind: "$resources",
									$layout: {
										$items: [{
											$bind: "dashboardName",
										}, {
											$bind: "title",
										}, {
											$bind: "description"
										}]
									}
								}]
							}
						}]
					}
				}]
			}
		},
		$lookup: {
			$copy: "$query"
		},
		$select: {
			$copy: "$query"
		}
	}
};
