// syracuse-ui/lib/tests/pages/ui-d3-graph/gantt.js
"use strict";
var _data = require("syracuse-ui/lib/tests/pages/ui-d3-graph/data/tasks");
exports.scenarios = {
	load: function() {
		return [{
			$title: "Scenario 1",
			run: function() {
				this.page.refreshData(_data);
				var field = this.page.boundFields["ui-d3-gantt"];
				if (field.$isEditMode) {

					// this.assert(fields && fields.length == 1 && fields[0] == syra_quality.domQuery.field.focus());
				}
			}
		}];
	}
};


exports.load = function(page, finish) {
	var $props = page.$prototype.$properties;
	$props["ui-d3-gantt"] = {
		$title: "ui-d3-gantt",
		$type: "application/x-gantt"
	};

	page.refreshData = function(data) {
		delete(this.dataset.$properties = this.dataset.$properties || {})["ui-d3-gantt"];
		delete this.dataset["ui-d3-gantt"];
		this.applyChange({
			$properties: {
				"ui-d3-gantt": data && data.$data.meta ? data.$data.meta : {}
			},
			"ui-d3-gantt": data && data.$data.rawData ? data.$data.rawData : null
		});
	};

	syra_item.ensureMediaTypes(page.$prototype, function() {
		page.addItem(page.leftCol, {
			$isEditMode: true,
			$bind: "ui-d3-gantt"
		});
		finish(page);
	});

};
