// syracuse-ui/lib/site/aside/devToolsPanel.js
"use strict";
var List = require('syracuse-ui/lib/scroll/list').List;
var OverPanel = require('syracuse-ui/lib/over/overPanel');
var classicActions = require('syracuse-ui/lib/fusion/core/client/sap/srvactions');

var _list, _popup;
var _classicFunctionLinks = ["GMAINT", "LCKSRC", "VERSYMB", "XXSEATRT", "ARECSRC", "RECHMESS", "LECTRACE", "EXETRT"];
var _classicPageLinks = ["SRV_ADXEXECUTE", "SRV_CALCULATRICE", "SRV_DEBUGGER", "LCL_TOGGLE_LOG"];
var _menuItemsClassicPage = {};
var _menuItemsClassicFunc = {};

exports.openByShortCut = function(page, field) {
	page.externalAdapter.onFieldClickPicker({
		field: field || page,
		pickerType: "devTools",
		doEvent: function() {
			exports.openPopup(syra_site.btns.devTools, page, field);
		}
	});
};

exports.openPopup = function(menuItem, page, item) {
	if (menuItem && menuItem.$item && menuItem.$item.$classicFunction) {
		syra_fusion.executeFusionFunc({
			"urlSeg": {
				"fusionParams": {
					"functionCode": menuItem.$item.$classicFunction
				}
			}
		}, syra_fusion.activatedBook);
	} else {
		if (!_popup) {
			_list = new List({
				css: "s-list-primary"
			});
			_list.addItem(syra_site, {
				$bind: "site_help_center",
				$category: "link",
				$skin: "s-list-primary-btn-default"
			}, "s-list-primary-bottom-sep");
			if (page && page.isFusionPage) {
				addClassicFunctionLinks(syra_site);
				addClassicPageLinks(page);
			}
			_popup = syra_over.openPopup(syra_site, {
				content: syra_site,
				slot: _list.scrollSlot,
				picker: menuItem.domItem,
				$isAutoClose: false,
				resizable: null,
				dragable: null,
				position: {
					my: "left top",
					at: "left bottom",
					of: menuItem.domItem
				},
				onresize: function(maxHeight) {
					_list && _list.resize(maxHeight);
				},
				close: function() {
					_list && _list.dispose();
					exports.cleanupLinks(true);
					_list = _popup = null;
				}
			});
		} else {
			exports.dispose();
		}
	}
	return true;
};

function addClassicPageLinks(page) {
	var i, len = _classicPageLinks.length;
	for (i = 0; i < len; i++) {
		_menuItemsClassicPage[_classicPageLinks[i]] = _list.addItem(page, {
			$bind: classicActions[_classicPageLinks[i].substr(0, 4) === "SRV_" ? "getSrvAction" : "getLocalAction"](_classicPageLinks[i]),
			$category: "link",
			$skin: "s-list-primary-btn-default"
		}, i != 0 ? null : "s-list-primary-top-sep");
	}
}

function addClassicFunctionLinks(page) {
	var i, len = _classicFunctionLinks.length;
	for (i = 0; i < len; i++) {
		_menuItemsClassicFunc[_classicFunctionLinks[i]] = _list.addItem(page, {
			$bind: "site_devTools",
			$category: "link",
			$title: _classicFunctionLinks[i],
			$classicFunction: _classicFunctionLinks[i],
			$skin: "s-list-primary-btn-default"
		});
	}
}
exports.cleanupLinks = function(dispose) {
	var arr, i, len, item;
	if (_popup || dispose) {
		arr = Object.keys(_menuItemsClassicPage);
		len = arr.length;
		for (i = 0; i < len; i++) {
			if ((item = _menuItemsClassicPage[arr[i]])) {
				syra_dom.remove(item.layoutSlot);
				item = null;
			}
		}
		_menuItemsClassicPage = {};
		arr = Object.keys(_menuItemsClassicFunc);
		len = arr.length;
		for (i = 0; i < len; i++) {
			if ((item = _menuItemsClassicFunc[arr[i]])) {
				if (!dispose) {
					item.disable(true);
				} else {
					item.dispose();
					item = null;
				}
			}
		}
		if (dispose) {
			_menuItemsClassicFunc = {};
		}
	}
};

exports.restoreLinks = function(page) {
	var arr, i, len, item;
	if (_popup) {
		arr = Object.keys(_menuItemsClassicFunc);
		len = arr.length;
		if (len) {
			for (i = 0; i < len; i++) {
				if ((item = _menuItemsClassicFunc[arr[i]])) {
					item.disable(false);
				}
			}
		} else {
			addClassicFunctionLinks(syra_site);
		}
		arr = Object.keys(_menuItemsClassicPage);
		len = arr.length;
		if (!len) {
			addClassicPageLinks(page);
		}
	}
};

exports.dispose = function() {
	_popup && _popup.close();
};
