// syracuse-ui/lib/site/aside/eventListener.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function EventListener(){
}

exports.EventListener = helpers.defineClass(EventListener, null, {
    load: function(){
        this._bindGlobalEvent();
        this._bindMenuClick();
        this._bindFieldEvents();
        this._bindSearchFacetEvents();
        this._bindResize();
        this._bindListEvents();
        this._bindChoiceFieldList();
        this._bindBox();
    },
    _bindGlobalEvent: function(){
        document.site.$$layoutSlot.bind("click", function(event){
            if (!document.site.authorPage) {
                var availableTarget = false;
                if (document.site.mainPage && document.site.mainPage.isFusionPage) {
                    var $$target = $(event.target);
                    ["data-s-menu", "data-s-field", "data-s-picker"].some(function(selector){
                        if ((event.target.getAttribute(selector) != null) ||
                        $$target.closest("[" + selector + "]").length > 0) 
                            return availableTarget;
                    });
                    if (!availableTarget) {
                        if ($$target.closest("[data-s-fusion-page]").length > 0) {
                            return document.site.mainPage.externalAdapter.onGlobalClick(event);
                        }
                    }
                }
            }
        });
    },
    _bindResize: function(){
        $(window).bind("resize", function(event){
            document.site.resize();
        });
    },
    _bindMenuClick: function(){
        document.site.$$layoutSlot.delegate("a[data-s-menu]", "click", function(event){
            try {
                var $$menu = $(this);
                var menu = document.controller.findItem($$menu);
                if (!menu) {
                    throw new Error("menu not found"); // to manage by
                    // diagnose
                }
                if (!menu.$isDisabled) {
                    menu.click();
                }
            } 
            catch (error) {
                document.site.onError(error);
                return false;
            }
            return false;
        });
    },
    bindAuthoringEvent: function(bind){
        if (bind) {
            document.site.$$layoutSlot.delegate(".s-aw-content-tree", "mousedow.authoring click.authoring", function(event){
                document.site.authorPage.treeContent.onNodeEvent(event);
                return false;
            }).delegate(".s-aw-tree-item-opener-children", "click.authoring", function(event){
                document.site.authorPage.treeContent.onNodeEvent(event);
                return false;
            }).delegate(".s-aw-tree-item-check", "click.authoring", function(event){
                document.site.authorPage.treeContent.onNodeEvent(event);
                return true;
            }).delegate(".s-aw-tree-item-title", "mouseenter.authoring mouseleave.authoring click.authoring", function(event){
                document.site.authorPage.treeContent.onNodeEvent(event);
                return false;
            });
        }
        else {
            document.site.$$layoutSlot.undelegate(".authoring");
        }
        
    },
    _bindFieldEvents: function(){
    
        document.site.$$layoutSlot.delegate("textarea[data-s-field],input[data-s-field],select[data-s-field]", "click change keydown keypress keyup focusin focusout", function(event){
            var field = document.controller.findField($(event.target));
            if (field) {
                var doEvent = true;
                if (event.type == "click") {
                    document.site.closePopups();
                }
                if (event.type == "click" && document.site.authorPage) {
                    doEvent = document.site.authorPage.onClickItem(field, this);
                }
                if (doEvent) {
                    field.onFieldInputEvent(event);
                    event.stopPropagation();
                }
            }
        }).delegate("a[data-s-field-binary]", "click focusin focusout", function(event){
            var field = document.controller.findField($(event.target));
            if (field) {
                field.onFieldInputEvent(event);
                if (event.type != "click") {
                    event.stopPropagation();
                    return false;
                }
            }
        }).delegate("a.s-icon-link", "click", function(event){
            var field = document.controller.findField($(event.target));
            if (field) {
                field.onFieldInputEvent(event);
                event.stopPropagation();
            }
            return false;
        }).delegate("[data-s-picker]", "click", function(event){
            var $$picker = $(event.currentTarget);
            var article = document.controller.findArticle($$picker);
            document.site.closePopups();
            if (article) {
                var pickerId = event.currentTarget.getAttribute("data-s-picker");
                if (!article.onClickPicker || article.onClickPicker(pickerId, event)) {
                    switch (pickerId) {
                        case "s-bar-dockMode":
                        case "s-bar-collapse":
                        case "s-bar-prev":
                        case "s-bar-next":
                        case "s-bar-title":
                            var isFusionBar = event.currentTarget.className.indexOf("s-fusion-bar") >= 0;
                            article[isFusionBar ? "fusionBar" : "menuBar"].onClickPicker(pickerId, event);
                            break;
                        default:
                            var field = document.controller.findField($$picker);
                            if (field) {
                                field.onClickPicker(event.currentTarget);
                            }
                            break;
                    }
                }
            }
            return false;
        }).delegate("[data-s-field-item]", "mouseenter mouseleave", function(event){
            var field = document.controller.findField($(event.target));
            if (field) {
                field.onFieldMouseEvent(event);
                //event.stopPropagation();
            }
        });
    },
    _bindListEvents: function(){
        document.site.$$layoutSlot.delegate("a[data-s-sort]", "click", function(event){
            var $$link = $(this);
            document.site.closePopups();
            var list = document.controller.findArticle($$link);
            list.sortCapability.onSortClick(event, $$link.attr("data-s-sort"));
            return false;
        }).delegate("a.s-alphatab-link", "click", function(){
            var $$link = $(this);
            document.site.closePopups();
            document.controller.findArticle($$link).sortCapability.onSortAlphaTab($$link);
            return false;
        }).delegate("a[data-s-cardview-opener]", "click", function(event){
            var $$link = $(this);
            document.site.closePopups();
            var article = document.controller.findArticle($$link);
            article.builder.cardview.onOpenerClick($$link, article);
            return false;
        }).delegate("a[data-s-page]", "click", function(event){
            var $$link = $(this);
            document.site.closePopups();
            document.controller.findArticle($$link).pagingCapability.onPageClick($$link);
            return false;
        }).delegate(".s-list-selector-row", "click mouseenter mouseleave", function(event){
            var list = document.controller.findArticle($(this)).list;
            if (event.type == "click") {
                document.site.closePopups();
                list._store.selector.onSelectRecords(event);
            }
            else {
                list._store.selector.onSelectMouseEvent(event);
            }
            return false;
        }).delegate(".s-list-selector,.s-list-selector-all", "click", function(event){
            var $$link = $(this);
            document.site.closePopups();
            var article = document.controller.findArticle($$link);
            event.stopPropagation();
            (article.list ? article.list : article)._store.selector.onSelectRecords(event);
        }).delegate("a.s-field-diagnose-link", "click", function(event){
            var $$link = $(event.currentTarget);
            document.site.closePopups();
            var field = document.controller.findField($$link);
            if (field) {
                field.focus();
            }
            event.stopPropagation();
        });
    },
    _bindSearchFacetEvents: function(){
        document.site.$$layoutSlot.delegate("input.s-search-facet-member-check", "change", function(){
            var $$input = $(this);
            document.controller.findArticle($$input).onInputMemberChange($$input);
            return false;
        });
    },
    _bindChoiceFieldList: function(){
        document.site.$$layoutSlot.delegate("a[data-s-cb-index]", "click", function(){
            var field = document.controller.findField($(this));
            if (field) {
                field.builder.onChoiceItemClick(parseInt(this.getAttribute("data-s-cb-index"), 10));
            }
            return false;
        });
    },
    _bindBox: function(){
        document.site.$$layoutSlot.delegate("a[data-s-box]", "click", function(event, data){
            var $$link = $(this);
            var box = document.controller.findBox($$link);
            if (box) {
                var doEvent = true;
                if (document.site.authorPage) {
                    doEvent = document.site.authorPage.onClickItem(box, this);
                }
                if (doEvent) {
                    if (box._tabTitle) {
                        var isTabSelected = box.layoutParent.tabOpened == box;
                        if (isTabSelected) {
                            box.page.externalAdapter.onBoxClick({
                                box: box,
                                event: event,
                                tabIdx: $$link.index()
                            });
                        }
                        else {
                            if (!document.site.DDAuthoring) {
                                var open = !box.$item.$opened;
                                var isFirstTime = !box.loaded;
                                box.page.externalAdapter.onBoxToggle({
                                    nativeEvent: event,
                                    nativeEvenData: data,
                                    box: box,
                                    isTabSelected: isTabSelected,
                                    open: open,
                                    isFirstTime: isFirstTime,
                                    tabIdx: $$link.index(),
                                    doEvent: function(){
                                        if (!isTabSelected) {
                                            box.openBox(open);
                                        }
                                        document.site.resize(true);
                                    }
                                });
                                return false;
                            }
                        }
                    }
                    else {
                        if (box.$item.$isBoxCollapsable) {
                            document.site.onBeforClick();
                            box.expandBox();
                        }
                    }
                }
            }
            return false;
        });
        document.site.$$layoutSlot.delegate("a[data-s-box-picker]", "click", function(event){
            document.controller.findBox($(this)).doPicker(this.getAttribute("data-s-box-picker"));
            return false
        });
    },
    dispose: function(){
        document.site.$$layoutSlot.unbind();
        document.site.$$layoutSlot.undelegate();
        $(window).unbind("resize");
    }
});
