// syracuse-ui/lib/site/aside/eventListener.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var keyboardShortCut = require('syracuse-ui/lib/site/aside/keyboardShortCut');

function EventListener(){
}


exports.EventListener = helpers.defineClass(EventListener, null, {
    load: function(){
        this._bindGlobalEvent();
        //this._bindShortCut();
        this._bindMenuClick();
        this._bindFieldEvents();
        this._bindSearchFacetEvents();
        this._bindResize();
        this._bindListEvents();
        this._bindChoiceFieldList();
        this._bindBox();
		this.tabKeyDownEvt = null;
		this.keysDown = [];
    },
    _bindGlobalEvent: function(){
        document.site.$$layoutSlot.bind("click", function(event){
            if (!document.site.authorPage) {
                var availableTarget = false;
                document.site.toggleTopPanel(false, event);
                document.site.jobsViewer && document.site.jobsViewer.toggle(false, event);
                if (document.site.mainPage && document.site.mainPage.isFusionPage) {
                    var $$target = $(event.target);
                    ["data-s-menu", "data-s-field", "data-s-picker"].some(function(selector){
                        if ((event.target.getAttribute(selector) != null) ||
                        $$target.closest("[" + selector + "]").length > 0) 
                            return availableTarget;
                    });
                    if (!availableTarget) {
                        if ($$target.closest("[data-s-fusion-page]").length > 0) {
                            return document.site.mainPage.externalAdapter.onGlobalClick(event);
                        }
                    }
                }
            }
		
        })
    },
	_bindShortCut: function(){
		var self = this;
		document.site.$$layoutSlot.delegate("*","keydown keyup",function(event){
			var elem = document.site;
			if( event.type == "keydown" ){
				self.keysDown.push(event.keyCode);
				if( event.keyCode == 9){ // fire tab keydown
					self.tabKeyDownEvt = event;
					return false;
				}else if( self.keysDown.length!==0 && self.keysDown.indexOf(9)!==-1 &&  event ){
					// manage short cut
					return keyboardShortCut.applyKeyboardShortCut(document.site,elem.page.$pageCategory === "fusion",self.keysDown);
				}
			}
			if( event.type == "keyup" ){
				delete self.keysDown[self.keysDown.indexOf(event.keyCode)];
				if(  event.keyCode == 9 ){
					var target = event.target;
					event = self.tabKeyDownEvt;
					if( elem.mainPage.$pageCategory !== "fusion" )
						self._focusOnNextFocusableField(event ? event : null);
				}
			}
		});
	},
    _bindResize: function(){
        $(window).bind("resize", function(event){
            document.site.resize();
        });
    },
    _bindMenuClick: function(){
        document.site.$$layoutSlot.delegate("a[data-s-menu]", "click", function(event){
            try {
                var $$menu = $(this);
                var menu = document.controller.findItem($$menu);
                if (!menu) {
                    throw new Error("menu not found"); // to manage by
                    // diagnose
                }
                if (!menu.$isDisabled) {
                    menu.click();
				}
            } 
            catch (error) {
                document.site.onError(error);
                return false;
            }
            return false;
        });
    },
    bindAuthoringEvent: function(bind){
        if (bind) {
            document.site.$$layoutSlot.delegate(".s-aw-content-tree", "mousedow.authoring click.authoring", function(event){
                document.site.authorPage.treeContent.onNodeEvent(event);
                return false;
            }).delegate(".s-aw-tree-item-opener-children", "click.authoring", function(event){
                document.site.authorPage.treeContent.onNodeEvent(event);
                return false;
            }).delegate(".s-aw-tree-item-check", "click.authoring", function(event){
                document.site.authorPage.treeContent.onNodeEvent(event);
                return true;
            }).delegate(".s-aw-tree-item-title", "mouseenter.authoring mouseleave.authoring click.authoring", function(event){
                document.site.authorPage.treeContent.onNodeEvent(event);
                return false;
            });
        }
        else {
            document.site.$$layoutSlot.undelegate(".authoring");
        }
        
    },
	_setFocus: function(f){
		if( f.parents(":hidden").length==0 && !f.attr("disabled") && f.is(':visible') && (f[0].localName !== 'a' || f.attr("href")) ){
			f.focus();
			console.log("focus "+f.html());
			return true ;
		}
		return false;
	},
	_focusOnNextFocusableField : function(event){
		var focused = false;
		var fields = $('#s-site-body').find('input,textarea,select,a');
		// get focus element
		var target = event ? event.target : null;
		var isShift = event ?  event.shiftKey : false;
		var index = target ? fields.index(target) : 0;
		if( index > -1 ){
			if( target )
				target.blur();
			if( isShift ){
				index--;
				while( (index) > 0 && !focused ) {
					var f = fields.eq(index);
					focused = this._setFocus(f);
					index--;
				}
			}else{
				index++;
				while( (index) < fields.length && !focused ) {
					var f = fields.eq(index);
					focused = this._setFocus(f);
					index++;
				}
			}
		}
		return false;
	},
    _bindFieldEvents: function(){
		var self = this; // manage key tab +
        document.site.$$layoutSlot.delegate("textarea[data-s-field],input[data-s-field],select[data-s-field]", "click change keydown keypress keyup focusin focusout", function(event){
            var field = document.controller.findField($(event.target));
            if (field) {
                if (document.site.authorPage && !document.site.authorPage.authorizeEvent(field)) {
                    return;
                }
                var doEvent = true;
                if (event.type == "click") {
                    document.site.closePopups();
                }
                if (event.type == "click" && document.site.authorPage) {
                    doEvent = document.site.authorPage.onClickItem(field, this);
                }

				/*
				 * not enouygh validate to uncomment
				 * if( event && event.keyCode === 9 ){
					if( event.type === "keydown"  ){
						doEvent = false;
					}else if(  event.type === "keyup"  ){
						event.type = "keydown";
					}
				}*/
                if (doEvent) {
                    field.onFieldInputEvent(event);
                }
                event.stopPropagation();
            }
        }).delegate("a[data-s-field-binary]", "click focusin focusout", function(event){
            var field = document.controller.findField($(event.target));
            if (field) {
                if (document.site.authorPage && !document.site.authorPage.authorizeEvent(field)) {
                    return;
                }
                field.onFieldInputEvent(event);
                if (event.type != "click") {
                    event.stopPropagation();
                    return false;
                }
            }
        }).delegate("a.s-icon-link", "click", function(event){
            var field = document.controller.findField($(event.target));
            if (field) {
                if (document.site.authorPage && !document.site.authorPage.authorizeEvent(field)) {
                    return;
                }
                field.onFieldInputEvent(event);
                event.stopPropagation();
            }
            return false;
        }).delegate("[data-s-picker]", "click", function(event){
            var $$picker = $(event.currentTarget);
            var article = document.controller.findArticle($$picker);
            document.site.onBeforClick();
            if (article) {
                if (document.site.authorPage && !document.site.authorPage.authorizeEvent(article)) {
                    return;
                }
                var pickerId = event.currentTarget.getAttribute("data-s-picker");
                if (!article.onClickPicker || article.onClickPicker(pickerId, event)) {
                    switch (pickerId) {
                        case "s-bar-dockMode":
                        case "s-bar-collapse":
                        case "s-bar-title":
                            var isFusionBar = event.currentTarget.className.indexOf("s-fusion-bar") >= 0;
                            article[isFusionBar ? "fusionBar" : "menuBar"].onClickPicker(pickerId, event);
                            break;
                        default:
                            var field = document.controller.findField($$picker);
                            if (field) {
                                field.onClickPicker(event.currentTarget);
                            }
                            break;
                    }
                }
            }
            return false;
        }).delegate("[data-s-field-item]", "mouseenter mouseleave", function(event){
            var field = document.controller.findField($(event.target));
            if (field) {
                field.onFieldMouseEvent(event);
                //event.stopPropagation();
            }
        });
    },
    _bindListEvents: function(){
    
        document.site.$$layoutSlot.delegate("a[data-s-sort]", "click", function(event){
            document.site.onBeforClick();
            var $$link = $(this);
            var list = document.controller.findArticle($$link);
            if (!document.site.authorPage || document.site.authorPage.authorizeEvent(list)) {
                list.sortCapability.onSortClick(event, $$link.attr("data-s-sort"));
            }
            return false;
        }).delegate("a.s-alphatab-link", "click", function(){
            document.site.onBeforClick();
            var $$link = $(this);
            var list = document.controller.findArticle($$link);
            if (!document.site.authorPage || document.site.authorPage.authorizeEvent(list)) {
                list.sortCapability.onSortAlphaTab($$link);
            }
            return false;
        }).delegate("a.s-cardview-opener,a.s-cardview-opener-all", "click", function(event){
            document.site.onBeforClick();
            var list = document.controller.findArticle($(this));
            list.builder.onCardViewOpenerClick(this);
            return false;
        }).delegate("a[data-s-page]", "click", function(event){
            document.site.onBeforClick();
            var $$link = $(this);
            var list = document.controller.findArticle($$link);
            if (!document.site.authorPage || document.site.authorPage.authorizeEvent(list)) {
                list.pagingCapability.onPageClick($$link);
            }
            return false;
        }).delegate(".s-list-selector-row", "click mouseenter mouseleave", function(event){
            var list = document.controller.findArticle($(this)).list;
            if (event.type == "click") {
                document.site.onBeforClick();
                if (list) {
                    list.selector.onSelectRecords(event);
                }
            }
            else {
                if (list) {
                    list.selector.onSelectMouseEvent(event);
                }
            }
            return false;
        }).delegate(".s-list-selector,.s-list-selector-all,.s-list-row-index-selector", "click", function(event){
            var $$link = $(this);
            document.site.onBeforClick();
            var article = document.controller.findArticle($$link);
            event.stopPropagation();
            (article.list ? article.list : article).selector.onSelectRecords(event);
        }).delegate("a.s-field-diagnose-link", "click", function(event){
            var $$link = $(event.currentTarget);
            document.site.onBeforClick();
            var field = document.controller.findField($$link);
            if (field) {
                field.focus();
            }
            event.stopPropagation();
        }).delegate("a.s-list-tree-node-picker", "click", function(event){
            document.site.onBeforClick();
            var article = document.controller.findArticle($(this));
            event.stopPropagation();
            article.list.treeDecorator.onNodeEvent(event);
            return false;
        });
    },
    _bindSearchFacetEvents: function(){
        document.site.$$layoutSlot.delegate("input.s-search-facet-member-check", "change", function(){
            var $$input = $(this);
            document.controller.findArticle($$input).onInputMemberChange($$input);
            return false;
        });
    },
    _bindChoiceFieldList: function(){
        document.site.$$layoutSlot.delegate("a[data-s-choice]", "click", function(){
            var field = document.controller.findField($(this));
            if (field) {
                field.builder.onChoiceItemClick(parseInt(this.getAttribute("data-s-choice"), 10));
            }
            return false;
        });
    },
    _bindBox: function(){
        document.site.$$layoutSlot.delegate("a[data-s-box]", "click", function(event, data){
            var $$link = $(this);
            var box = document.controller.findBox($$link);
            if (box) {
                var doEvent = true;
                if (document.site.authorPage) {
                    doEvent = document.site.authorPage.onClickItem(box, this);
                }
                if (doEvent) {
                    if (box.tabTitle) {
                        var isTabSelected = box.layoutParent.tabOpened == box;
                        if (isTabSelected) {
                            box.page.externalAdapter.onBoxClick({
                                box: box,
                                event: event,
                                tabIdx: $$link.index()
                            });
                        }
                        else {
                            if (!document.site.DDAuthoring) {
                                var open = !box.$item.$opened;
                                var isFirstTime = !box.loaded;
                                box.page.externalAdapter.onBoxToggle({
                                    nativeEvent: event,
                                    nativeEvenData: data,
                                    box: box,
                                    isTabSelected: isTabSelected,
                                    open: open,
                                    isFirstTime: isFirstTime,
                                    tabIdx: $$link.index(),
                                    doEvent: function(){
                                        if (!isTabSelected) {
                                            box.openBox(open);
                                        }
                                        document.site.resize(true);
                                    }
                                });
                                return false;
                            }
                        }
                    }
                    else {
                        if (box.$item.$isBoxCollapsable) {
                            document.site.onBeforClick();
                            box._expandBody();
                        }
                    }
                }
                else {
                    event.stopPropagation();
                }
            }
            return false;
        });
        document.site.$$layoutSlot.delegate("a[data-s-box-picker]", "click", function(event){
            setTimeout(function(){
                document.controller.findBox($(event.target)).doPicker(event.target.getAttribute("data-s-box-picker"));
            }, 20);
            return false
        });
    },
    dispose: function(){
        document.site.$$layoutSlot.unbind();
        document.site.$$layoutSlot.undelegate();
        $(window).unbind("resize");
    }
});
