// syracuse-ui/lib/site/aside/externalAdapter.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function ExternalAdapter(){
}

exports.ExternalAdapter = helpers.defineClass(ExternalAdapter, null, {
    onGlobalClick: function(event){
        return true;
    },
    onFieldNotifyChange: function(options){
        options.doEvent();
    },
    onFieldClickPicker: function(options){
        options.doEvent();
    },
    onSelectRecordEvent: function(options){
        options.doEvent();
    },
    onSortClickEvent: function(options){
        options.doEvent();
    },
    onBoxToggle: function(options){
        options.doEvent();
    },
    onBoxClick: function(options){
        return true;
    },
    onFieldEvent: function(options){
        if (document.site.onFieldFocusIn && options.event.type == "focusin") {
            document.site.onFieldFocusIn(options.field);
        }
        options.doEvent();
    },
    releaseMainPage: function(options){
        options.doEvent();
    },
    onMenuItemClick: function(options){
        options.doEvent();
    },
    setDataBind: function(field, $resources, record, metaData){
        return true;
    },
    setDataBindArray: function(field, $resources, record, metaData){
        return true;
    },
    createField: function(){
        return null;
    },
    setPickerValue: function(options){
        options.doStatements(options);
    },
    notifyFldMandatoryErr: function(){
        return true;
    },
    applyDesignMetaExt: function(){
        return false;
    }
});
