// syracuse-ui/lib/site/aside/formula.js
"use strict";
/*
 Copyright (c)2012 Sage.
 http://wwww.sage.com
 */

function _getPrototype(localize) {
	return {
		"$type": "application/json",
		"$title": localize.formula_page_title,
		"$x3Facet": "message",
		"$properties": {
			"formula": {
				"$type": "application/x-string",
				"$X3Name": "AAFORMULA_formula",
				"$X3Fmt": "K{-1,-1,-1,32}:50X",
				"$maxLength": 1022,
				"$groupName": "B",
			},
			"help": {
				"$type": "application/x-vignette",
				"$title": localize.formula_fld_help,
				"$isHidden": true,
				"$location": {
					"$type": "html"
				}
			},
			"fl": {
				"$isFieldPager": false,
				"$type": "application/x-array",
				"$item": {
					"$properties": {
						"fl1": {
							"$title": "Key",
							"$type": "application/x-string",
							"$capabilities": "",
							"$maxLength": 10
						},
						"fl2": {
							"$title": "Parent",
							"$type": "application/x-string",
							"$capabilities": "",
							"$maxLength": 10
						},
						"fl3": {
							"$title": "Title",
							"$type": "application/x-string",
							"$capabilities": "",
							"$maxLength": 10
						},
						"fl4": {
							"$title": "Open",
							"$type": "application/x-boolean",
							"$capabilities": "",
							"$maxLength": 10
						}
					}
				},
				"$treeview": {
					"$mode": "parentKey",
					"$bindings": {
						"$clientId": "fl2",
						"$parent": "fl1",
						"$description": "fl3",
						"$open": "fl4"
					}
				}
			},
			"operator": {
				"$isFieldPager": false,
				"$type": "application/x-array",
				"$item": {
					"$properties": {
						"operator1": {
							"$title": "Key",
							"$type": "application/x-string",
							"$capabilities": "",
							"$maxLength": 10
						},
						"operator2": {
							"$title": "Parent",
							"$type": "application/x-string",
							"$capabilities": "",
							"$maxLength": 10
						},
						"operator3": {
							"$title": "Title",
							"$type": "application/x-string",
							"$capabilities": "",
							"$maxLength": 10
						},
						"operator4": {
							"$title": "Open",
							"$type": "application/x-boolean",
							"$capabilities": "",
							"$maxLength": 10
						}
					}
				},
				"$treeview": {
					"$mode": "parentKey",
					"$bindings": {
						"$clientId": "operator2",
						"$parent": "operator1",
						"$description": "operator3",
						"$open": "operator4"
					}
				}
			},
			"fc": {
				"$isFieldPager": false,
				"$type": "application/x-array",
				"$item": {
					"$properties": {
						"fc1": {
							"$title": "Key",
							"$type": "application/x-string",
							"$capabilities": "",
							"$maxLength": 10
						},
						"fc2": {
							"$title": "Parent",
							"$type": "application/x-string",
							"$capabilities": "",
							"$maxLength": 10
						},
						"fc3": {
							"$title": "Title",
							"$type": "application/x-string",
							"$capabilities": "",
							"$maxLength": 10
						},
						"fc4": {
							"$title": "Open",
							"$type": "application/x-boolean",
							"$capabilities": "",
							"$maxLength": 10
						}
					}
				},
				"$treeview": {
					"$mode": "parentKey",
					"$bindings": {
						"$clientId": "fc2",
						"$parent": "fc1",
						"$description": "fc3",
						"$open": "fc4"
					}
				}
			}
		},
		"$links": {
			"$verify": {
				"$title": localize.formula_act_verify,
				"$act": 1070
			}
		}
	};
}

function _getFieldSection(localize) {
	return {
		$category: "section",
		$title: localize.formula_blck_fields,
		$isBoxCollapsable: true,
		$layout: {
			$items: [{
				$selectMode: "row",
				$bind: "fl",
				$isEditMode: false,
				$isDockedHeight: true,
				$isMenuRecordHidden: true,
				$isPagerHidden: false,
				$isPopupMenuHidden: true,
				$isQuickDesignerDisabled: true,
				$isTopLabelAlignment: true,
				$layout: {
					$items: [{
						$bind: "fl1"
					}, {
						$bind: "fl2"
					}, {
						$bind: "fl3"
					}, {
						$bind: "fl4"
					}]
				}

			}]
		}
	};
}

function _getFunctionSection(localize) {
	return {
		$category: "section",
		$title: localize.formula_blck_functions,
		$isBoxCollapsable: true,
		$layout: {
			$items: [{
				$bind: "fc",
				$selectMode: "row",
				$format: "grid",
				$isDockedHeight: true,
				$isEditMode: false,
				$isMenuRecordHidden: true,
				$isPagerHidden: false,
				$isPopupMenuHidden: true,
				$isQuickDesignerDisabled: true,
				$isTopLabelAlignment: true,
				$layout: {
					$items: [{
						$bind: "fc1"
					}, {
						$bind: "fc2"
					}, {
						$bind: "fc3"
					}, {
						$bind: "fc4"
					}]
				}
			}]
		}
	};
}

function _getOperatorsSection(localize) {
	return {
		$category: "section",
		$title: localize.formula_blck_operators,
		$isBoxCollapsable: true,
		$layoutCss: "s-formula",
		$layout: {
			$items: [{
				$bind: "operator",
				$selectMode: "row",
				$fixedBodyHeight: "150px",
				$format: "cards",
				$addTreeLevel: true,
				$skin: "s-htree",
				$isEditMode: false,
				$isMenuRecordHidden: true,
				$isPagerHidden: false,
				$isPopupMenuHidden: true,
				$isQuickDesignerDisabled: true,
				$isTopLabelAlignment: true,
				$layout: {
					$items: [{
						$bind: "operator1"
					}, {
						$bind: "operator2"
					}, {
						$bind: "operator3"
					}, {
						$bind: "operator4"
					}]
				}

			}]
		}
	};
}

function _getFormuleSection(localize) {
	return {
		$category: "section",
		$title: localize.formula_fld_formula,
		$isBoxCollapsable: true,
		$layout: {
			$items: [{
				$bind: "formula",
				$isTitleHidden: true,
				$isEditMode: true,
				$rows: 4
			}, {
				$skin: "s-page-link",
				$category: "link",
				$bind: "$verify"
			}]
		}
	};
}

function _getHelpSection() {
	return {
		$category: "section",
		$isTitleHidden: true,
		$layout: {
			$items: [{
				$rows: 10,
				$isBoxCollapsable: true,
				$opened: false,
				$isDockedHeight: true,
				$bind: "help"
			}]
		}
	};
}

function _getArticle(localize) {
	return {
		$layout: {
			$items: [{
				$category: "section",
				$layout: {
					$layoutType: "row",
					$widths: "25,50,25",
					$items: [_getFieldSection(localize), {
							$layoutType: "stack",
							$items: [_getOperatorsSection(localize), _getFormuleSection(localize), _getHelpSection(localize)]
						},
						_getFunctionSection(localize)
					]
				}
			}]

		}
	};
}

exports.getFormulaPage = function() {
	var localize = syra_site.localize;
	var $prototype = _getPrototype(localize);
	$prototype.$article = _getArticle(localize);
	return $prototype;
};
