// syracuse-ui/lib/site/aside/formula.js
"use strict";
/*
 Copyright (c)2012 Sage.
 http://wwww.sage.com
 */

function _getPrototype(localize) {
	return {
		"$type": "application/json",
		"$title": localize.formula_page_title,
		"$x3Facet": "message",
		"$properties": {
			"AA1": {
				"$type": "application/x-string",
				"$X3Name": "AAFORMULA_AA1",
				"$X3Fmt": "K{-1,-1,-1,32}:50X",
				"$maxLength": 1022,
				"$groupName": "B",
			},
			"showHelp": {
				"$type": "application/x-boolean",
				"$title": localize.formula_fld_help,
				"$format": "$check",
			},
			"help": {
				"$type": "application/x-vignette",
				"$title": localize.formula_fld_help,
				"$isHidden": true,
				"$location": {
					"$type": "html",
					$url: "http://www.sage.fr/fr"
				}
			},
			"bA": {
				"$isFieldPager": false,
				"$type": "application/x-array",
				"$item": {
					"$properties": {
						"bA0": {
							"$title": "Key",
							"$type": "application/x-string",
							"$capabilities": "",
							"$maxLength": 10
						},
						"bA1": {
							"$title": "Parent",
							"$type": "application/x-string",
							"$capabilities": "",
							"$maxLength": 10
						},
						"bA2": {
							"$title": "Title",
							"$type": "application/x-string",
							"$capabilities": "",
							"$maxLength": 10
						},
						"bA3": {
							"$title": "Open",
							"$type": "application/x-boolean",
							"$capabilities": "",
							"$maxLength": 10
						}
					}
				},
				"$treeview": {
					"$mode": "parentKey",
					"$bindings": {
						"$id": "bA0",
						"$parent": "bA1",
						"$description": "bA2",
						"$open": "bA3"
					}
				}
			},
			"bB": {
				"$isFieldPager": false,
				"$type": "application/x-array",
				"$item": {
					"$properties": {
						"bB0": {
							"$title": "Key",
							"$type": "application/x-string",
							"$capabilities": "",
							"$maxLength": 10
						},
						"bB1": {
							"$title": "Parent",
							"$type": "application/x-string",
							"$capabilities": "",
							"$maxLength": 10
						},
						"bB2": {
							"$title": "Title",
							"$type": "application/x-string",
							"$capabilities": "",
							"$maxLength": 10
						},
						"bB3": {
							"$title": "Open",
							"$type": "application/x-boolean",
							"$capabilities": "",
							"$maxLength": 10
						}
					}
				},
				"$treeview": {
					"$mode": "parentKey",
					"$bindings": {
						"$id": "bB0",
						"$parent": "bB1",
						"$description": "bB2",
						"$open": "bB3"
					}
				}
			},
			"bC": {
				"$isFieldPager": false,
				"$type": "application/x-array",
				"$item": {
					"$properties": {
						"bC0": {
							"$title": "Key",
							"$type": "application/x-string",
							"$capabilities": "",
							"$maxLength": 10
						},
						"bC1": {
							"$title": "Parent",
							"$type": "application/x-string",
							"$capabilities": "",
							"$maxLength": 10
						},
						"bC2": {
							"$title": "Title",
							"$type": "application/x-string",
							"$capabilities": "",
							"$maxLength": 10
						},
						"bC3": {
							"$title": "Open",
							"$type": "application/x-boolean",
							"$capabilities": "",
							"$maxLength": 10
						}
					}
				},
				"$treeview": {
					"$mode": "parentKey",
					"$bindings": {
						"$id": "bC0",
						"$parent": "bC1",
						"$description": "bC2",
						"$open": "bC3"
					}
				}
			}
		},
		"$links": {
			"$verify": {
				"$title": localize.formula_act_verify,
				"$act": 1070
			}
		}
	};
}

function _getFieldSection(localize) {
	return {
		$category: "section",
		$title: localize.formula_blck_fields,
		$isBoxCollapsable: true,
		$layout: {
			$items: [{
				$selectMode: "row",
				$bind: "bA",
				$isEditMode: false,
				$isPageHeightDocked: true,
				$isMenuRecordHidden: true,
				$isPagerHidden: false,
				$isPopupMenuHidden: true,
				$isQuickDesignerEnabled: false,
				$isTopLabelAlignment: true,
				$layout: {
					$items: [{
						$bind: "bA0"
					}, {
						$bind: "bA1"
					}, {
						$bind: "bA2"
					}, {
						$bind: "bA3"
					}]
				}

			}]
		}
	};
}

function _getFunctionSection(localize) {
	return {
		$category: "section",
		$title: localize.formula_blck_functions,
		$isBoxCollapsable: true,
		$layout: {
			$items: [{
				$bind: "bC",
				$selectMode: "row",
				$format: "grid",
				$isPageHeightDocked: true,
				$isEditMode: false,
				$isMenuRecordHidden: true,
				$isPagerHidden: false,
				$isPopupMenuHidden: true,
				$isQuickDesignerEnabled: false,
				$isTopLabelAlignment: true,
				$layout: {
					$items: [{
						$bind: "bC0"
					}, {
						$bind: "bC1"
					}, {
						$bind: "bC2"
					}, {
						$bind: "bC3"
					}]
				}
			}]
		}
	};
}

function _getOperatorsSection(localize) {
	return {
		$category: "section",
		$title: localize.formula_blck_operators,
		$isBoxCollapsable: true,
		$layout: {
			$items: [{
				$bind: "bB",
				$selectMode: "row",
				$ssfixedBodyHeight: "150px",
				$format: "cards",
				$addTreeLevel: true,
				$skin: "s-htree",
				$isEditMode: false,
				$isMenuRecordHidden: true,
				$isPagerHidden: false,
				$isPopupMenuHidden: true,
				$isQuickDesignerEnabled: false,
				$isTopLabelAlignment: true,
				$layout: {
					$items: [{
						$bind: "bB0"
					}, {
						$bind: "bB1"
					}, {
						$bind: "bB2"
					}, {
						$bind: "bB3"
					}]
				}

			}]
		}
	};
}

function _getFormuleSection(localize) {
	return {
		$category: "section",
		$title: localize.formula_fld_formula,
		$isBoxCollapsable: true,
		$layout: {
			$items: [{
				$bind: "AA1",
				$isTitleHidden: true,
				$isEditMode: true,
				$rows: 4
			}, {
				$skin: "s-page-link",
				$category: "link",
				$bind: "$verify"
			}]
		}
	};
}

function _getHelpSection() {
	return {
		$category: "section",
		$isTitleHidden: true,
		$layout: {
			$items: [{
				$bind: "showHelp",
				$isEditMode: true
			}, {
				$rows: 10,
				$isBoxCollapsable: true,
				$opened: false,
				$isPageHeightDocked: true,
				$bind: "help"
			}]
		}
	};
}

function _getArticle(localize) {
	return {
		$layout: {
			$items: [{
				$category: "section",
				$layout: {
					$layoutType: "row",
					$widths: "25,50,25",
					$items: [_getFieldSection(localize), {
							$layoutType: "stack",
							$items: [_getOperatorsSection(localize), _getFormuleSection(localize), _getHelpSection(localize)]
						},
						_getFunctionSection(localize)
					]
				}
			}]

		}
	};
}

exports.getFormulaPage = function() {
	var localize = document.site.localize;
	var $prototype = _getPrototype(localize);
	$prototype.$article = _getArticle(localize);
	return $prototype;
};
