// syracuse-ui/lib/site/aside/legalPage.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var DesktopPage = require("syracuse-ui/lib/page/desktopPage").DesktopPage;

function LegalPage() {}

exports.LegalPage = helpers.defineClass(LegalPage, DesktopPage, {
	loadBox: function($prototype) {
		this.layoutSlot = document.createElement("div");
		this.localize = document.site.localize;

		this.$prototype = $prototype;
		this.$prototype.$properties.$copyright = {
			$type: "application/x-string",
			$isReadOnly: true
		};
		this.$prototype.$links = this.$prototype.$links || {};
		this.$prototype.$links.$eula = {
			$title: document.site.localize.legalEula,
			$url: ""
		};

		this.$item = {
			$category: "page",
			$title: document.site.localize.legalPageTitle,
			$layout: {
				$category: "section",
				$isTitleHidden: true,
				$layout: {
					$items: [{
						$isTitleHidden: true,
						$layout: {
							$layoutType: "row",
							$items: [{
								$bind: "$copyright",
								$isTitleHidden: true
							}, {
								$category: "link",
								$bind: "$eula"
							}]
						}

					}, {
						$format: "cards",
						$isTitleHidden: true,
						$isPagerHidden: true,
						$isQuickDesignerDisabled: true,
						$isRowIndexVisible: false,
						$selectByRowIndex: false,
						$isMenuRecordHidden: true,
						$bind: "$resources",
						$layout: {
							$items: [{
								$layoutType: "stack",
								$items: [{
									$bind: "product"
								}, {
									$bind: "productVersion"
								}, {
									$bind: "serial"
								}, {
									$bind: "licenseeName"
								}, {
									$bind: "licenseeRegistrationNumber"
								}, {
									$bind: "valEnd"
								}, {
									$bind: "resellerName"
								}, {
									$bind: "resellerRegistrationNumber"
								}]
							}]
						}

					}]
				}

			}
		};
		DesktopPage.prototype.loadBox.call(this);
	},
	applyChange: function(newData) {
		// append copyright data
		newData.$copyright = document.site.localize.legalCopyright;

		DesktopPage.prototype.applyChange.call(this, newData);
	},
	onMenuClick: function(menuItem) {
		if (menuItem.$bind == "$eula") {
			alert("EULA downloading not set yet");
		}
	}

});
