// syracuse-ui/lib/site/aside/siteTopPanel.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Article = require("syracuse-ui/lib/article/article").Article;
var locale = require('syracuse-core/lib/locale');

function SiteTopPanel(){
}

exports.SiteTopPanel = helpers.defineClass(SiteTopPanel, Article, {
    loadBox: function(){
        this._localize = locale.resources(module)();
        document.itemFactory.initializeItem(this, {});
        document.site.setZIndex((this.$$item = this.$$container)[0]);
        Article.prototype.loadBox.call(this);
        if (document.site.userProfile) {
            document.site.userProfile.ensureFormBlock();
            if (document.site.userProfile.$$item) {
                if (document.site.userProfile.$$item.parent().length == 0) {
                    this.$$topUserProfile.empty().append(document.site.userProfile.$$item);
                }
            }
        }
        this.$$container[0].style.height = (this.$$body.outerHeight() + this.$$body.offset().top) + "px";
    },
    drawBox: function(){
        this.$$body = $("<div id='s-site-top-pn-body'/>").appendTo(this.$$item);
        this.renderTopBar();
        this.renderNavigationBar();
        this.renderAboutBar();
    },
    renderTopLink: function(options){
        var dom = document.createElement("a");
        dom.className = "s-top-link";
        dom.setAttribute("href", "#");
        var $$link = $(dom);
        dom = document.createElement("div");
        dom.className = "s-top-link-icon " + options.$icon;
        $$link.append(dom);
        
        dom = document.createElement("div");
        dom.className = "s-top-link-title";
        $$link.append($(dom).text(options.$title).append($("<div class='s-top-link-desc'/>").text(options.$desc)));
        $$link.bind("click", options.$click).appendTo(this.$$topLinks);
    },
    renderTopBar: function(){
        var self = this;
        self.$$topBar = $("<div id='s-site-top-pn-top-bar'/>").appendTo(self.$$body);
        self.$$topUserProfile = $("<div id='s-site-top-pn-user-profile'/>").appendTo(self.$$topBar);
        
        self.$$topLinks = $("<div id='s-site-user-top-pn-links'/>").appendTo(self.$$topBar);
        self.renderTopLink({
            $icon: "s-top-link-bookmark-open",
            $title: self._localize.dtp_bookmarkLinkTitle,
            $desc: self._localize.dtp_bookmarkLinkDesc,
            $click: function(){
                return false;
            }
        });
        self.renderTopLink({
            $icon: "s-top-link-aw-open",
            $title: self._localize.dtp_authoringLinkTitle,
            $desc: self._localize.dtp_authoringLinkDesc,
            $click: function(){
                document.site.toggleAuthoring(true);
                return false;
            }
        });
    },
    renderNavigationPanel: function(navItem){
        var div = document.createElement("div");
        div.className = "s-site-top-pn-nav-bar-cell " + navItem.$css;
        navItem.$$cell = $(div).appendTo(this.$$navigationBar);
        
        navItem.domTitle = document.createElement("div");
        navItem.domTitle.className = "s-site-top-pn-nav-title";
        navItem.$$cell[0].appendChild(navItem.domTitle).textContent = navItem.$title;
        
        div = document.createElement("div");
        div.className = "s-site-top-pn-nav-item-body";
        navItem.links = document.itemFactory.load($(div).appendTo(navItem.$$cell), {
            $category: "links",
            $isBindDisabled: navItem.$isBindDisabled || false,
            $links: navItem.$links,
            $skin: "s-site-top-pn-nav-bar-menus"
        }, this);
        navItem.$$cell.bind("mouseleave mouseenter", function(event){
            navItem.links.$$item.find("a").toggleClass("s-site-top-pn-nav-bar-cell-over", event.type == "mouseenter");
            return false;
        });
    },
    renderNavigationBar: function(){
        var self = this;
        self.$$navigationBar = $("<div id='s-site-top-pn-nav-bar'/>").appendTo(self.$$body);
        Object.keys(self.navItems = {
            drafts: {
                $title: self._localize.dtp_drafts,
                $css: "s-site-top-pn-drafts"
            },
            historic: {
                $title: self._localize.dtp_historic,
                $isBindDisabled: true,
                $css: "s-site-top-pn-historic",
                $links: {
                    "banks": {
                        "$title": self._localize.dtp_historic_links_banks
                    },
                    "properties": {
                        "$title": self._localize.dtp_historic_links_properties
                    },
                    "newAsset": {
                        "$title": self._localize.dtp_historic_links_newAsset
                    },
                    "listAsset": {
                        "$title": self._localize.dtp_historic_listAsset
                    },
                    "account": {
                        "$title": self._localize.dtp_historic_account
                    }
                }
            },
            bookmarks: {
                $title: self._localize.dtp_bookmarks,
                $css: "s-site-top-pn-bookmarks"
            },
            dashboards: {
                $title: self._localize.dtp_dashboards,
                $css: "s-site-top-pn-dashboards",
                $isBindDisabled: true,
                $links: self.$prototype.$navigation.$dashboard
            }
        }).forEach(function(id){
            self.renderNavigationPanel(self.navItems[id]);
        });
    },
    renderAboutBar: function(){
        this.$$aboutBar = $("<div id='s-site-top-pn-about-bar'/>").appendTo(this.$$body);
        document.itemFactory.load(this.$$aboutBar, {
            $category: "block",
            $title: this._localize.dtp_about,
            $skin: "s-site-top-pn-about-bar",
            $layout: {
                $items: [{
                
                    $category: "links",
                    $isBindDisabled: true,
                    $links: this.$prototype.$navigation.$about,
                    $skin: "s-site-top-pn-about-menus"
                }]
            }
        }, this);
    },
    closePanel: function(){
        var $$item = document.site.userProfile ? document.site.userProfile.$$item : null;
        if ($$item) {
            $$item.detach();
        }
        this.$$container.hide().empty();
        document.controller.disposeObject(this);
    },
    dispose: function(){
        var self = this;
        if (self.navItems) {
            Object.keys(self.navItems).forEach(function(prop){
                var navItem = self.navItems[prop];
                if (navItem && navItem.$$cell) {
                    navItem.$$cell.unbind();
                }
            });
        }
        Article.prototype.dispose.call(this);
    }
});
