// @sage/syracuse-lib/src/upload/helpers.ts
"use strict";
import * as config from 'config';
import * as ez from 'ez-streams';
import * as htmlTransforms from '@sage/html-sanitizer/src/transforms';

var allowedTypes = config.upload && config.upload.allowedTypes;
if (!allowedTypes) {
	if (config.hosting && config.hosting.multiTenant) throw new Error("config.upload.allowedTypes not configured");
	else allowedTypes = /./;
}
if (!Array.isArray(allowedTypes)) allowedTypes = [allowedTypes];

var rejectors=[];

export function allowContentType(contentType) {
	// Get rid of options
	contentType = contentType && contentType.split(";")[0];
	return allowedTypes.some(function(re) {
		return re.test(contentType);
	});
};

function MediaTypeError(_, reader, contentType) {
	Error.call(this);
	// // consume stream to prevent hanging on client side, it would be better to abort request but need further investigation
	while (reader.read(_)); //nop
	console.error("upload rejected: contentType=" + contentType);
	this.message = "unauthorized media type";
	this.$httpStatus = 415;
}

import * as mmm from 'mmmagic';
var magic = new mmm.Magic(mmm.MAGIC_MIME_TYPE);

export function detectMediaType(buf, _) {
	if (typeof(buf) === "string") {
		buf = new Buffer(buf, "binary");
	}
	return magic.detect(buf, _);
}

export function sanitizeReader(_, reader, contentType) {
	// if application/octet-stream skip content-type check to find a better one with mmmagic
	if (contentType && contentType !== "application/octet-stream" && !exports.allowContentType(contentType)) throw new MediaTypeError(_, reader, contentType);
	reader.setEncoding(null);
	reader = reader.peekable();
	var buf = reader.peek(_);
	if (buf == null)
		return reader;
	var detected = detectMediaType(buf, _);
	if (!exports.allowContentType(detected)) throw new MediaTypeError(_, reader, detected);
	if (/html/i.test(contentType) || /html/i.test(detected)) {
		reader = reader.map(function(_, buf) {
			return buf.toString('binary');
		}).transform(htmlTransforms.escaper({
			preserve: true,
			profile: 'html'
		}));
	} else {
		if (/svg/i.test(contentType) || /svg/i.test(detected)) {
			reader = reader.map(function(_, buf) {
				return buf.toString('binary');
			}).transform(htmlTransforms.escaper({
				preserve: false, // preserve might result in svg errors and failure to display
				profile: 'svg'
			}));
		}
	}
	return reader;
};
exports.registerRejector = function (cb) {
	if (rejectors.indexOf(cb) < 0){
		rejectors.push(cb);
	}
}
