// syracuse-ui/lib/field/array/tools/filler.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var RecordArticle = require("syracuse-ui/lib/field/array/recordArticle").RecordArticle;
var _queryFilters = require("syracuse-ui/lib/field/array/tools/queryFilters");
var _parser = require('syracuse-sdata/lib/parser/parser');
var _formatApi = require('syracuse-ui/lib/field/formatApi');


exports.notifyClientSave = function(list, savedDelta) {
	if (savedDelta && savedDelta.$uuid) {
		var dataRecord = exports.findDataRecord(list, savedDelta.$uuid);
		var delta = {};
		var $binds = Object.keys(list.$prototype.$item.$properties);
		for (var ii = 0, jj = $binds.length; ii < jj; ii++) {
			var $bind = $binds[ii];
			if (savedDelta[$bind] !== undefined) {
				delta[$bind] = savedDelta[$bind];
			}
		}
		delta.$uuid = savedDelta.$uuid;
		delta.$key = savedDelta.$key;
		if (dataRecord) {
			syra_site.deltaManager.applyObjectDelta(list.page, dataRecord, delta);
			if (list.recordsMap[savedDelta.$uuid]) {
				list.recordsMap[savedDelta.$uuid].applyChange(delta);
			}
		} else {
			var dataset = list.ensureDataSet();
			delta.$index = dataset.length;
			delta = [delta];
			syra_site.deltaManager.applyPageArrayDelta(list.page, dataset, delta);
			list.setDataBind(delta, list.articleParent.ensureDataSet());
		}
	}
};

exports.trigger = function(list, options, $location) {
	if (!list.disposed && !list.isReloading) {
		options = options || {};
		var params = {
			where: list.$where,
			select: list.$select
		};
		if (options) {
			if (options.$filters) {
				params.filter = (options.$filters == "$noFilters") ? "null" : options.$filters;
				_queryFilters.onFetch(list, list.paramsFilter = params.filter);
			} else {
				if (list.paramsFilter) {
					params.filter = list.paramsFilter;
				}
			}
			params.startIndex = options.startIndex;
			params.count = list.pagging.getItemsPerPage(options.$itemsPerPage);
			if (options.filter != null) {
				params.startIndex = 1;
				params.where = list.$where = options.filter != "" ? options.filter : null;
			}
			if (options.queryBindings) {
				if (options.queryBindings.length == 0) {
					var $binds = Object.keys(list.$fields);
					for (var ii = 0, jj = $binds.length; ii < jj; ii++) {
						options.queryBindings.push($binds[ii]);
					}
				}
				list.select = options.queryBindings.map(function(binding) {
					return binding.replace(/\./g, '/');
				}).join(',');
			}
		}
		list.capabilityMaker.sorter.setFetchParams(list, params, options);
		var fetchOptions = {
			params: params
		};
		if (list.isClientFetch || list.isClientSortEnable) {
			var $keys = Object.keys(options);
			exports.fetchClientRecordSet(list, fetchOptions, $keys.length == 1 && ($keys == "$itemsPerPage" || $keys == "startIndex"));
		} else {
			fetchOptions.field = list;
			fetchOptions.$location = $location;
			if (params.where !== undefined) {
				list.page.$prototype.$where = params.where;
			}
			if (params.filter !== undefined) {
				list.page.$prototype.$filter = params.filter;
			}
			list.page.fetch(fetchOptions, function(data, response, requestUrl) {
				list.page.isMainPage && syra_site.history.update(list.page, requestUrl);
				if (list.page.dataset) {
					delete list.page.dataset[list.$item.$bind];
					var steps = ["$first", "$previous", "$next", "$last"];
					var $links = list.page.dataset.$links || {};
					for (var ii = steps.length - 1; ii >= 0; ii--) {
						delete $links[steps[ii]];
						delete list.page.$menus[steps[ii]];
					}
				}
				if (list.page.$properties) {
					delete list.page.$properties[list.$item.$bind];
				}
				list.page.startChange();
				list.page.applyChange(data, response, requestUrl);
				list.page.endChange();
			});
		}
		list.articleParent.onFetchHandler && list.articleParent.onFetchHandler(fetchOptions);
	}
};

exports.applyFetchOptions = function(list, options, pagingChange) {
	list.ensureDataSet();
	if (!pagingChange || (list.clientDataset == list.dataset)) {
		list.clientDataset = [];
		for (var ii = 0, jj = list.dataset.length; ii < jj; ii++) {
			var dataRecord = list.dataset[ii];
			list.clientDataset.push(dataRecord);
		}
	}
	if (options && options.params && !pagingChange) {
		if (options.params.where) {
			var exp = _parser.Parser.parse(options.params.where);
			var currentDataset = list.clientDataset;
			list.clientDataset = [];
			for (var ii = 0, jj = currentDataset.length; ii < jj; ii++) {
				var record = currentDataset[ii];
				var ff = record[exp.children[0].value];
				if (ff != null) {
					var ok = false;
					if (typeof(ff) == "object") {
						// quick workaround ($title). to improve for generic cases
						ok = _formatApi.compare(ff.$title, exp.children[1].value, "string", exp.value.code);
					} else {
						if (typeof(exp.children[1].value) == "string") {
							ok = _formatApi.compare(ff, exp.children[1].value, "string", exp.value.code);
						} else {
							// filter date
							if (typeof(exp.children[1].value) == "object" && list.$fields[exp.children[0].value] && list.$fields[exp.children[0].value].$type == "application/x-date") {
								ok = _formatApi.compare(ff, exp.children[1].value, "date", exp.value.code);
							} else {
								// case of integer (application/x-choice)
								if (typeof(ff) == "number") {
									ok = _formatApi.compare(ff, exp.children[1].value, "number", exp.value.code);
								} else {
									ok = false;
								}
							}
						}
					}
					ok && list.clientDataset.push(record);
				}
			}
		}
		if (options.params.orderBy !== undefined) {
			list.capabilityMaker.sorter.applyFetchOptions(list, options.params);
		}
	}
};
exports.fetchClientRecordSet = function(list, options, pagingChange) {
	if (options && options.params) {
		list.page._isDataChanging = true;
		exports.applyFetchOptions(list, options, pagingChange);
		if (options.params.count || options.params.startIndex) {
			list.pagging.ensureClientRange(options.params.count, options.params.startIndex);
		}
		list.clientFetchOptions = options;
		list.fillList(list.clientDataset, list.articleParent.dataset);
		list.page._isDataChanging = false;
		syra_site.ensureArticleVisibility(list, true);
		list.resizeArticle(true);
	}
};

exports.ensureClientDataset = function(list) {
	if (!list.clientDataset) {
		list.clientDataset = list.ensureDataSet();
	}
	return list.clientDataset;
};

exports.getClientDataSetIndex = function(list, $serverIndex) {
	exports.ensureClientDataset(list);
	for (var ii = 0, jj = list.clientDataset.length; ii < jj; ii++) {
		if (list.clientDataset[ii].$serverIndex == $serverIndex) {
			return ii;
		}
	}
	return -1;
};

function _applyFullDelta(list, dataRecordSet, dataIndex, updatedMap) {
	var oldMap = list.recordsMap;
	list.recordsMap = {};
	var oldRecords = list.records;
	list.records = [];
	var lastRecord, recordIndex = 0;
	for (var ii = dataIndex.start; ii <= dataIndex.last; ii++) {
		var dataRecord = dataRecordSet[ii - 1];
		var $index = dataRecord.$serverIndex;
		if ($index === undefined) {
			$index = dataRecord.$index;
		}
		if ($index === undefined) {
			$index = (ii - 1);
		}
		var sourceDataset = list.dataset[$index];
		if (sourceDataset) {
			var record = oldMap[dataRecord.$uuid];
			if (record) {
				record.ensureDataSet(dataRecord.$uuid);
				var updatedRecord = updatedMap && updatedMap[dataRecord.$uuid];
				list.recordsMap[dataRecord.$uuid] = record;
				list.records.push(record);
				record.applyChange(updatedRecord || dataRecord);
				record.$serverIndex = sourceDataset.$serverIndex;
				if (oldRecords.indexOf(record) != recordIndex) {
					if (lastRecord) {
						list.builder.record_reorder(record, lastRecord, true);

					} else {
						list.builder.record_reorder(record, null, false);
					}
				}
			} else {
				record = exports.addRecord(list, {
					dataRecord: sourceDataset,
					$recordIndex: recordIndex
				});
			}
			lastRecord = record;
			recordIndex++;
		}
		delete oldMap[dataRecord.$uuid];
	}
	var uiids = Object.keys(oldMap);
	for (var ii = 0, jj = uiids.length; ii < jj; ii++) {
		exports.removeRecord(oldMap[uiids[ii]], true);
		delete oldMap[uiids[ii]];
	}
};

function _applyPartialDelta(list, deltaSet, dataIndex) {
	for (var $serverIndex = 0, jj = deltaSet.length; $serverIndex < jj; $serverIndex++) {
		var deltaRecord = deltaSet[$serverIndex];
		var record = list.recordsMap[deltaRecord.$uuid];
		if (record) {
			var index = parseInt(deltaRecord.$index, 10) + 1;
			if (deltaRecord.$isDeleted || !(dataIndex.start <= index && index <= dataIndex.last)) {
				exports.removeRecord(record, true, true);
			} else {
				var recordIndex = null,
					prevIndex = null;
				if (deltaRecord.$index !== undefined && record.$serverIndex != deltaRecord.$index) {
					prevIndex = list.records.indexOf(record);
					recordIndex = deltaRecord.$index - dataIndex.start;
				}
				record.ensureDataSet(record.$uuid);
				record.applyChange(deltaRecord);
				if (prevIndex) { //not process if list.clientFetchOptions
					list.records.splice(prevIndex, 1);
					list.records.splice(recordIndex, 0, record);
					var nextRecord;
					if ((recordIndex + 1) < list.records.length) {
						nextRecord = list.records[recordIndex + 1];
					}
					list.builder.record_reorder(record, nextRecord, nextRecord ? false : true);
				}
			}
		} else {
			if (!deltaRecord.$isDeleted) {
				var index = parseInt(deltaRecord.$index, 10) + 1;
				if (dataIndex.start <= index && index <= dataIndex.last) {
					exports.addRecord(list, {
						dataRecord: list.dataset[deltaRecord.$index],
						$recordIndex: deltaRecord.$index
					});
				}
			}
		}
	}
};


exports.fill = function(list, dataRecordSet, parentDataRecord, isDelta) {
	if (isDelta) {
		var updatedMap = {};
		if (!list.isSingList && dataRecordSet) {
			for (var ii = dataRecordSet.length - 1; ii >= 0; ii--) {
				updatedMap[dataRecordSet[ii].$uuid] = dataRecordSet[ii];
			}
		}
		var partialDelta = list.page.$isPartialDelta || (dataRecordSet.length == 1 && dataRecordSet[0].$index !== undefined);
		if (list.clientFetchOptions) {
			exports.applyFetchOptions(list, list.clientFetchOptions);
			if (list.pagging.$startIndex !== undefined && list.clientDataset.length < list.pagging.$startIndex) {
				list.pagging.ensureClientRange();
				list.clientFetchOptions.params.startIndex = list.pagging.$startIndex;
			}
			var dataIndex = list.pagging.getParseIndexes();
			_applyFullDelta(list, list.clientDataset, dataIndex, updatedMap);
		} else {
			list.clientDataset = list.ensureDataSet();
			var dataIndex = list.pagging.getParseIndexes();
			if (partialDelta) {
				_applyPartialDelta(list, dataRecordSet, dataIndex);
			} else {
				_applyFullDelta(list, dataRecordSet, dataIndex);
			}
		}
	} else {
		list.clientDataset = list.isSingList ? list.ensureDataSet() : dataRecordSet;
		list.body.style.display = "none";
		exports.removeRecords(list, !(list.clientDataset && list.clientDataset.length > 0));
		if (list.clientDataset) {
			var start, last;
			//warn: list.onClientFetch =true and list.isClientFetch=false for convergence left list
			if (list.isClientFetch) {
				if (list.pagging.$startIndex !== undefined && list.clientDataset.length < list.pagging.$startIndex) {
					list.pagging.ensureClientRange();
				}
				var dataIndex = list.pagging.getParseIndexes();
				start = dataIndex.start;
				last = dataIndex.last;
			} else {
				start = 1;
				last = list.clientDataset.length;
			}
			for (var ii = start; ii <= last; ii++) {
				var dataRecord = list.clientDataset[ii - 1];
				if (!dataRecord.$isDeleted) {
					exports.addRecord(list, {
						dataRecord: dataRecord
					});
				}
			}
		}
	}
	exports.validateDisplay(list);
};

function _disableAdd(list, isAddDisabled) {
	list._isAddDisabled = isAddDisabled;
	if (list.menuItems) {
		if (list.menuItems.$create) {
			list.menuItems.$create[0].disable(list._isAddDisabled);
		}
		if (list.menuItems.$select) {
			list.menuItems.$select[0].disable(list._isAddDisabled);
		}
	}
	if (list.$capability.insert) {
		for (var ii = 0, jj = list.records.length; ii < jj; ii++) {
			var record = list.records[ii];
			if (record.menuItems.$create) {
				record.menuItems.$create[0].disable(list._isAddDisabled);
			}
		}
	}
}

exports.validateDisplay = function(list) {
	if (!list.disableDisplayValidation) {
		var showAlt = true;
		list.greatestServerIndexRecord = null;
		for (var ii = 0, jj = list.records.length; ii < jj; ii++) {
			var record = list.records[ii];
			if (record.dataset.$index !== undefined) {
				record.dataset.$index = ii;
			}
			if (record.dataset.$serverIndex == undefined) {
				if (record.singleField) {
					record.dataset.$serverIndex = ii;
				} else {
					for (var mm = 0, kk = list.dataset.length; mm < kk; mm++) {
						if (list.dataset[mm] == record.dataset) {
							record.dataset.$serverIndex = mm;
							break;
						}
					}
				}
			}
			record.$serverIndex = record.dataset.$serverIndex;
			if (!list.greatestServerIndexRecord || list.greatestServerIndexRecord.$serverIndex < record.$serverIndex) {
				list.greatestServerIndexRecord = record;
			}
			record.setRowIndex && record.setRowIndex();
			if (list.$item.$alternateStyle) {
				syra_site.dom.toggleClass(record.domItem, "s-list-alt", showAlt = !showAlt);
				record.freezeRow && syra_site.dom.toggleClass(record.freezeRow, "s-list-alt", showAlt);
			}
			if (record.expandRowCard) {
				list.builder.cardDecorator.toggleRowCard(record, true);
			}
		}
		//_checkMaxItems
		if (list.$isEditMode) {
			if (list.$prototype.$minItems) {
				while (list.records.length < list.$prototype.$minItems) {
					exports.addRecord(list, {
						dataRecord: {}
					});
				}
			}
			if (list.$prototype.$maxItems) {
				if (list._isAddDisabled !== (list.$prototype.$maxItems <= list.dataset.length)) {
					_disableAdd(list, !list._isAddDisabled);
				}
			}
		}
	}
};

function _clearEmpty(list) {
	var empty = list.builder.emptyDataItem;
	if (empty) {
		if (empty.syraIsDisplay !== false) {
			empty.syraIsDisplay = false;
			empty.style.display = "none";
		} else {
			syra_site.dom.removeChild(empty);
			empty = null;
		}
	}
}

exports.addRecord = function(list, options) {
	if (list.isSingList) {
		options.dataRecord = {
			$singleField: options.dataRecord
		};
		options.$prototype = {
			$properties: {
				$singleField: list.$prototype.$item
			}
		};
	}
	_clearEmpty(list);
	if (options.dataRecord.$uuid === undefined) {
		options.dataRecord.$uuid = options.$recordIndex || list.records.length;
	}
	var record = (list.recordsMap[options.dataRecord.$uuid] = new RecordArticle());
	record.arrayLevel = "record";
	record.$uuid = options.dataRecord.$uuid;
	list.recordsMap[record.$uuid] = record;
	if (options.$recordIndex !== undefined) {
		record.insertBeforeRecord = list.records[options.$recordIndex];
		list.records.splice(options.$recordIndex, 0, record);
	} else {
		list.records.push(record);
	}
	record.garbage = list.garbage;
	if (options.dataRecord) {
		record.dataset = options.dataRecord;
		record.$serverIndex = options.dataRecord.$serverIndex;
	}
	record.$facet = list.$recordFacet;
	record.$isEditMode = list.$isEditMode && !list.treeDecorator;
	record.list = list;
	record.$prototype = options.$prototype || list.$prototype.$item;
	list.page.initializeNewItem(record, options.$item || {}, list);
	record.loadBox(null, options.isCreateAction);
	record.applyChange(list.ensureGlobalMetaRecord());
	options.dataRecord && record.applyChange(options.dataRecord);
	list.builder.onAfterAddRecord && list.builder.onAfterAddRecord(record);
	return record;
};


exports.removeRecord = function(record, removeDom, removeContext) {
	var list = record.list;
	if (record.treeNode && list.treeDecorator) {
		record.treeNode = null;
	}
	if (removeContext) {
		list.records.splice(list.records.indexOf(record), 1);
		delete list.recordsMap[record.$uuid];
	}
	list.removeItem(record, removeDom);
};

exports.removeRecords = function(list, addEmptySlot) {
	for (var ii = 0, jj = list.records.length; ii < jj; ii++) {
		var record = list.records[ii];
		delete list.recordsMap[record.$uuid];
		exports.removeRecord(record);
	}
	list.recordsMap = {};
	list.records = [];
	if (list.selector && list.selector.clear) {
		list.selector.clear();
	}
	list.emptyBody(addEmptySlot);
};


function _addEmptyItem(list, onFetch) {
	var empty = list.builder.emptyDataItem = document.createElement("div");
	empty.className = list.$skin + "-empty-slot";
	if (!list.isClientFetch && onFetch) {
		empty.textContent = syra_local.flWaitingData;
	} else {
		empty.textContent = list.builder.emptyMessage || list.$item.$noDataText || syra_local.flNoData;
	}
	list.body.appendChild(empty);
}

exports.emptyBody = function(list, addEmptySlot, onFetch) {
	syra_site.dom.empty(list.body);
	addEmptySlot && _addEmptyItem(list, onFetch);
};

function _notifySelectRecords(list, menuItem) {
	if (list.currentSelectRecords) {
		var sendBag = list.page.ensureSendBag(list);
		var dataRecords = list.ensureDataSet();
		var $uuids = Object.keys(list.currentSelectRecords);
		var hasNew, lastRecord;
		if (list.isSingList) {
			sendBag.dataRecords = dataRecords;
			for (var ii = 0, jj = $uuids.length; ii < jj; ii++) {
				var options = {
					isCreateAction: true,
					dataRecord: list.currentSelectRecords[$uuids[ii]].dataset
				};
				switch (list.$prototype.$item.$type) {
					case "application/x-choice":
						options.dataRecord = options.dataRecord.$uuid;
						break;
					case "application/x-reference":
						delete options.dataRecord.$serverIndex;
						delete options.dataRecord.$isSelected;
						break;
				}
				if (menuItem.$item.$variantItemKey) {
					var variantValue = options.dataRecord;
					(options.dataRecord = {})[menuItem.$item.$variantItemKey] = variantValue;
				}
				dataRecords.push(options.dataRecord);
				lastRecord = exports.addRecord(list, options);
				hasNew = true;
			}
		} else {
			var $newRecords = [];
			for (var ii = 0, jj = $uuids.length; ii < jj; ii++) {
				if (!exports.findDataRecord(list, $uuids[ii])) {
					$newRecords.push(list.currentSelectRecords[$uuids[ii]].dataset);
					hasNew = true;
				}
			}
			if (hasNew) {
				if ($newRecords.length == 1) {
					$newRecords[0].$index = list.dataset.length;
					sendBag.dataRecords = $newRecords;
				} else {
					sendBag.dataRecords = [];
					for (var ii = 0, jj = list.dataset.length; ii < jj; ii++) {
						sendBag.dataRecords.push({
							$uuid: list.dataset[ii].$uuid
						});
					}
					sendBag.dataRecords = sendBag.dataRecords.concat($newRecords);
				}
			}
		}
		if (hasNew) {
			exports.validateDisplay(list);
			list.page.notifyDataChange(list, sendBag.dataRecords);
			lastRecord && lastRecord.scrollToRecord();
		}
	}
}

exports.doSelectAction = function(list, menuItem) {
	list.currentSelectRecords = null;
	syra_site.dialogManager.openModal(list.boxParent, {
		article: list,
		$url: menuItem.$sourceUrl,
		onValidate: function() {
			_notifySelectRecords(list, menuItem);
		},
		onSelectRecord: function(selectedRecords) {
			list.currentSelectRecords = selectedRecords;
			return false;
		}
	});
};

exports.findRecordByServerIndex = function(list, $serverIndex) {
	for (var ii = 0, jj = list.records.length; ii < jj; ii++) {
		if (list.records[ii].$serverIndex == $serverIndex) {
			return list.records[ii];
		}
	}
	return null;
};

exports.findRecord = function(list, target) {
	while (target && target != document) {
		if (target.syraRecord !== undefined) {
			return list.recordsMap[target.syraRecord];
		}
		target = target.parentNode;
	}
	return null;
};

exports.findDataRecord = function(list, $uuid) {
	var found;
	if (list.dataset) {
		for (var ii = 0, jj = list.dataset.length; ii < jj; ii++) {
			var dataRecord = list.dataset[ii];
			if (dataRecord && dataRecord.$uuid == $uuid) {
				found = {
					dataRecord: dataRecord,
					dataRecordIndex: ii
				};
				break;
			}
		}
	}
	return found;
};

exports.getField = function(list, name, $serverIndex, select) {
	var fields, record = exports.findRecordByServerIndex(list, $serverIndex);
	if (record && !record.disposed) {
		if ((fields = record.boundFields[name]) && fields.length > 0) {
			return (list.builder.popupCardBuilder && list.builder.popupCardBuilder.findField(name, record)) || fields[0];
		}
	}
	return null;
};
exports.getDataValue = function(list, name, $serverIndex) {
	var dataRecordSet = list.ensureDataSet();
	var field;
	if ($serverIndex !== undefined) {
		if (field = exports.getField(list, name, $serverIndex)) {
			return field.getDataValue();
		} else {
			for (var ii = 0, jj = dataRecordSet.length; ii < jj; ii++) {
				if (dataRecordSet[ii].$serverIndex == $serverIndex) {
					var value = dataRecordSet[ii][name];
					return value === undefined ? null : value;
				}
			}
		}
	}
	return null;
};

exports.setFocus = function(list, select, name, $serverIndex) {
	if (list.builder.onBeforeListSetFocus) {
		list.builder.onBeforeListSetFocus(select, name, $serverIndex);
	}
	if (list.treeDecorator) {
		var record = exports.findRecordByServerIndex(list, $serverIndex);
		record && list.selector.selectRecord(record.$uuid, true);
		return true;
	}
	if (list.$capability.search) {
		list.searchCapability.start(list, list.dataset[$serverIndex].$uuid, name, $serverIndex);
	}
	var field = exports.getField(list, name, $serverIndex);
	if (field) {
		return field.focus(select);
	} else {
		if (list.isClientFetch && !exports.findRecordByServerIndex(list, $serverIndex)) {
			list.pagging.setPage($serverIndex);
			field = exports.getField(list, name, $serverIndex);
			if (field) {
				var record = exports.findRecordByServerIndex(list, $serverIndex);
				record && list.selector.selectRecord(record.$uuid, true);
				return field.focus(select);
			}
		}
	}
	return null;
};

exports.setFocusRecord = function(list, record) {
	list.page.setShortcutSpyerList(list);
	if (list.focusedRecord && list.focusedRecord != record) {
		syra_site.dom.toggleClass(list.focusedRecord.domItem, "s-record-focused", false);
		list.focusedRecord.freezeRow && syra_site.dom.toggleClass(list.focusedRecord.freezeRow, "s-record-focused", false);
	}
	list.focusedRecord = record;
	syra_site.dom.toggleClass(record.domItem, "s-record-focused", true);
	record.freezeRow && syra_site.dom.toggleClass(record.freezeRow, "s-record-focused", true);
};

exports.getFocusRecord = function(list, selected) {
	if (list.focusedRecord && !list.focusedRecord.disposed) {
		return list.focusedRecord;
	}
	var record = null;
	if (selected && list.selector.records) {
		if (list.selector.uuid != undefined) {
			record = list.selector.records[list.selector.uuid];
		} else {
			var keys = Object.keys(list.selector.records);
			if (keys.length > 0) {
				record = list.selector.records[keys[0]];
			}
		}
	}
	return record;
};

exports.toggleFakeRecord = function(list, show) {
	var dataRecords = list.ensureDataSet();
	if (show) {
		if (dataRecords.length == 0) {
			exports.ensureClientDataset(list);
			var options = {
				isCreateAction: true,
				$serverIndex: 0,
				dataRecord: {
					$isFakeRecord: true,
					$uuid: helpers.uuid.generate()
				}
			};
			dataRecords.splice(options.$serverIndex, 0, options.dataRecord);
			exports.addRecord(list, options);
		}
	} else {
		if (list.records.length == 1) {
			var record = list.records[0];
			if (record.dataset.$isFakeRecord) {
				if (list.clientDataset) {
					list.clientDataset.splice(list.clientDataset.indexOf(record.dataset), 1);
				}
				list.records.splice(list.records.indexOf(record), 1);
				dataRecords.splice(list.clientDataset.indexOf(record.dataset), 1);
				delete list.recordsMap[record.$uuid];
				exports.removeRecord(record, true);
				list.emptyBody(true);
			}
		}
	}
};
