// syracuse-ui/lib/field/array/tools/slotResizerDDAgent.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function SlotResizerDDAgent() {}

exports.SlotResizerDDAgent = helpers.defineClass(SlotResizerDDAgent, null, {
	start: function(dropItem) {
		this.builder = dropItem.builder;
		document.site.ddManager.ddAgent = this;
		if (dropItem.resizeCard) {
			this.resizeCard = true;
			this.slot = this.builder.cardDecorator.outCardSlot;
			this.separator = this.builder.cardDecorator.outCardSeparator;
			this.isLeftMove = this.builder.list.$item.$cardItem.$position == "left";
		} else {
			if (dropItem.resizeGraph) {
				this.resizeGraph = true;
				this.slot = this.builder.list.graphDecorator.graphSlot;
				this.separator = this.builder.list.graphDecorator.separator;
				this.isLeftMove = this.builder.list.$item.$graphPosition == "left";
			}
		}
		var offset = this.slot.getBoundingClientRect();
		this.left = offset.left;
		this.top = offset.top;
		this.width = offset.width;
		this.height = offset.height;
		this.right = offset.right;
		this.bottom = offset.bottom;
		this.startX = this.separator.getBoundingClientRect().left;
	},
	onDragMouseMove: function(target, event) {
		var cursor = "not-allowed";
		if ((event.pageY >= (this.top - 20)) && (event.pageY <= (this.bottom + 20))) {
			var newWidth = this.width + (this.isLeftMove ? (event.pageX - this.startX) : (this.startX - event.pageX));
			if (newWidth > 200) {
				this.slot.style.width = (this.newWidth = newWidth) + "px";
				cursor = "e-resize";
			}
		}
		document.site.body.style.cursor = cursor;
	},
	onDragMouseUp: function(target, event) {
		var percent = Math.floor(100 * (this.slot.clientWidth / this.slot.parentNode.clientWidth)) + "%";
		this.slot.style.width = percent;
		if (this.resizeCard) {
			this.builder.list.$item.$cardItem.$width == percent;
		} else {
			if (this.resizeGraph) {
				this.builder.list.$item.$graphWidth == percent;
			}
		}
		this.builder.list.notifyDesignerUpdate();
		this.builder.list.onWindowResize();
		document.site.body.style.cursor = "default";
	},
	dispose: function() {
		this.slot = this.separator = this.startX = this.builder = null;
	}
});
