// syracuse-ui/lib/field/array/tools/treeDecorator.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function TreeDecorator() {}

exports.TreeDecorator = helpers.defineClass(TreeDecorator, null, {
	buildTreeMap: function(dataset, partialDelta) {
		var $bindings = this.list.$prototype.$treeview.$bindings;
		var jj = dataset.length,
			extAdapter = this.list.page && this.list.page.externalAdapter,
			listItems = this.list.$prototype && this.list.$prototype.$item;
		if (!partialDelta) {
			this._uuidNodes = {};
			this._nodes = {};
			this.tree = [];
		}
		if (jj && listItems && extAdapter && extAdapter.isListRecordEmpty(listItems.$properties, dataset[jj - 1])) {
			// Unfortunately, in Convergence last line could be inconsistent.
			jj--;
		}
		for (var ii = 0; ii < jj; ii++) {
			var dataRecord = dataset[ii];
			var parentId = dataRecord[$bindings.$parent];
			var id = dataRecord[$bindings.$clientId] || "s-empty";
			var node;
			var isOpened;
			if (partialDelta) {
				node = this._uuidNodes[dataRecord.$uuid] || this._nodes[id];
				if (parentId === undefined && node) {
					parentId = node.parentId;
				}
				if ($bindings.$open && dataRecord[$bindings.$open] !== undefined) {
					isOpened = $bindings.$open ? (dataRecord[$bindings.$open] === true || dataRecord[$bindings.$open] == 2) : true;
				} else {
					isOpened = node ? node.isOpened : false;
				}
			} else {
				node = this._nodes[id];
				isOpened = $bindings.$open ? (dataRecord[$bindings.$open] === true || dataRecord[$bindings.$open] == 2) : true;
			}
			var appendToParent = false;
			if (node) {
				if (node.parentId == undefined) {
					var found = this.tree.indexOf(node);
					if (found >= 0) {
						this.tree.splice(found, 1);
						appendToParent = true;
					}
				} else {
					if (node.parentId != parentId) {
						var prevParent = this._nodes[node.parentId];
						if (prevParent) {
							prevParent.children.splice(prevParent.children.indexOf(node), 1);
						}
						appendToParent = true;
					}
				}
			} else {
				appendToParent = true;
				this._nodes[id] = node = {
					id: id,
					desc: dataRecord[$bindings.$description] || "",
					children: []
				};
			}
			if (appendToParent) {
				var parentNode = this._nodes[parentId];
				if (!parentNode) {
					this.tree.push(parentNode = this._nodes[parentId] = {
						id: parentId,
						children: []
					});
				}
				parentNode.children.push(node);
			}
			node.$uuid = dataRecord.$uuid;
			this._uuidNodes[node.$uuid] = node;
			node.parentId = parentId;
			node.isOpened = isOpened;
		}
	},
	onNodeSelected: function(uuid, selected) {
		var parendId = (this._uuidNodes[uuid] ? this._uuidNodes[uuid].parentId : null),
			parent = (parendId && parendId !== "") ? this._nodes[this._uuidNodes[uuid].parentId] : null,
			selectParent = selected,
			childCount;
		if (parent) {
			if (!selectParent) {
				if (parent.children && (childCount = parent.children.length)) {
					for (var ii = 0; ii < childCount; ii++) {
						if (this.list.dataset[parent.children[ii].$uuid] && this.list.dataset[parent.children[ii].$uuid].$isSelected) {
							selectParent = true;
							break;
						}
					}
				}
			}
			if (parent.$uuid !== undefined)
				this.list.selector.select(parent.$uuid, selectParent);
		}
	},
	load: function(list) {
		this.list = list;
		var $skin = this.list.$item.$treeSkin || "s-list-tree";
		this.cssNodeSelector = $skin + "-selector";
		this.cssNodeCell = $skin + "-cell";
		this.cssNode = $skin + "-node";
		this.cssNodePicker = this.cssNode + "-picker";
		this.cssNodeIcon = this.cssNode + "-icon";
		this.cssNodeDesc = this.cssNode + "-desc";
		this.cssNodeDescValue = this.cssNode + "-desc-value";
		this.cssNodeTitleCell = this.cssNode + "-title-cell";
		this.list.$isEditMode = false; // set false editMode for that that contains tree
		var $bindings = this.list.$prototype.$treeview.$bindings;
		var $datas = $bindings.$data || [];
		/*if (!$datas) {
         var $keys = Object.keys($bindings);
         for (var ii = 0, jj = $keys.length; ii < jj; ii++) {
         var $field = list.$fields[$bindings[$keys[ii]]];
         if ($field) {
         $field.$isHidden = true;
         }
         }
         }*/
		var $keys = Object.keys(list.$fields);
		this.$visibleFields = [];
		for (var ii = 0, jj = $keys.length; ii < jj; ii++) {
			var $field = list.$fields[$keys[ii]];
			$field.$isExcluded = $datas.indexOf($keys[ii]) < 0;
			if (!$field.$isHidden && !$field.$isExcluded) {
				this.$visibleFields.push($field);
			}
		}
	},
	findNode: function(record) {
		return this._nodes[record.dataset[this.list.$prototype.$treeview.$bindings.$clientId] || "s-empty"];
	},
	onNodeEvent: function(event) {
		var self = this;
		var record = self.list.findRecord(event.target);
		self.list.page.externalAdapter.onFieldClickPicker({
			field: record,
			pickerType: "openTreeNode",
			doEvent: function() {
				if (record && record.treeNode) {
					var node = self.findNode(record);
					node.isOpened = !node.isOpened;
					record.treeNode.picker.setAttribute("data-s-picker", "list-tree-picker");
					record.treeNode.picker.className = self.cssNodePicker + (node.isOpened ? " s-open" : " s-close");
					self.displayChildren(node, node.isOpened, node.children);
					self.list.resizeList();
				}
			}
		});
	},
	expendCollapseNode: function($uuid, expend) {
		var self = this;
		if ((this._uuidNodes[$uuid].isOpened && !expend) || (!this._uuidNodes[$uuid].isOpened && expend)) {
			this.list.recordsMap[$uuid].treeNode.picker.click();
			self.list.resizeList();
		}
	},
	expendCollapseAll: function(expend) {
		var records = this.list.records,
			node, self = this;
		for (var ii = 0, jj = records.length; ii < jj; ii++) {
			node = this.findNode(records[ii]);
			if (node) {
				if (expend === node.isOpened)
					continue;
				node.isOpened = expend;
				records[ii].treeNode.picker.setAttribute("data-s-picker", "list-tree-picker");
				records[ii].treeNode.picker.className = this.cssNodePicker + (node.isOpened ? " s-open" : " s-close");
				this.displayChildren(node, node.isOpened, node.children);
			}
		};
		setTimeout(function() {
			self.list.resizeList();
		}, 100);
	},
	getPreviousNextServerIndexVisible: function(recordIndex, next) {
		var list = this.list,
			index;
		for (var ii = 0, jj = this._orderedNodes.length; ii < jj; ii++) {
			if (this._orderedNodes[ii].$serverIndex == recordIndex) {
				index = ii;
				break;
			}
		}
		for (var ii = (next ? Math.min(index + 1, this._orderedNodes.length - 1) : Math.max(index - 1, 0)), jj = (next ? this._orderedNodes.length : -1);
			(next && ii < jj) || (!next && ii > jj);
			(next ? ii++ : ii--)) {
			if (this._uuidNodes[this._orderedNodes[ii].$uuid].isVisible) {
				return parseInt(this._orderedNodes[ii].$serverIndex, 10);
			}
		}
		return index;
	},
	displayChildren: function(parentNode, isParentOpened, children) {
		if (parentNode.childRoot) {
			parentNode.childRoot.style.display = isParentOpened ? "" : "none";
		}
		for (var ii = 0, jj = children.length; ii < jj; ii++) {
			var node = children[ii];
			this.list.recordsMap[node.$uuid].dataRow.style.display = (node.isVisible = isParentOpened) ? "" : "none";
			if (node.children.length) {
				this.displayChildren(node, node.isOpened && isParentOpened, node.children);
			}
		}
	},
	drawNodesOrder: function(parentNode, nodes, level) {
		var isCard = this.list.$item.$format == "cards";
		var skinLevel = this.list.$item.$addTreeLevel ? (" " + this.list.$skin + "-level-" + level) : null;
		var parentRecords = {};
		if (nodes) {
			for (var ii = 0, jj = nodes.length; ii < jj; ii++) {
				var node = nodes[ii];
				var hasChildren = node.children.length > 0;
				var record = (node.$uuid != undefined) ? this.list.recordsMap[node.$uuid] : null;
				if (record) {
					this._orderedNodes.push({
						$uuid: record.$uuid,
						$serverIndex: record.$serverIndex,
						$recordIndex: record.$recordIndex
					});
					if (isCard) {
						if (hasChildren) {
							parentRecords[node.$uuid] = record;
							if (!node.childRoot) {
								node.childRoot = document.createElement("div");
								node.childRoot.className = this.list.$skin + "-children";
								if (skinLevel) {
									node.childRoot.className += skinLevel;
								}
								record.domItem.appendChild(node.childRoot);
								node.childRoot.style.display = "none";
							}
						}
						if (parentNode && parentNode.childRoot) {
							parentNode.childRoot.appendChild(record.dataRow);
						} else {
							this.list.body.appendChild(record.dataRow);
						}
					} else {
						this.list.body.appendChild(record.dataRow);
					}
					record.treeNode.level = level;
					if (this.list.selector.appendToRecord) {
						if (this.list.selector.isMulti) {
							if (!record._selectorCheck) {
								record._selectorCheck = this.list.selector.appendToRecord(record.treeNode.selectorSlot);
								record.treeNode.selectorSlot.style.display = "";
							}
						} else {
							if (hasChildren) {
								if (record._selectorCheck) {
									document.site.dom.removeChild(record._selectorCheck);
									delete record._selectorCheck;
									record.treeNode.selectorSlot.style.display = "none";
								}
							} else {
								if (!record._selectorCheck) {
									record._selectorCheck = this.list.selector.appendToRecord(record.treeNode.selectorSlot);
									record.treeNode.selectorSlot.style.display = "";
								}
							}
						}
					}
					if (hasChildren) {
						record.treeNode.picker.setAttribute("data-s-picker", "list-tree-picker");
						record.treeNode.picker.className = this.cssNodePicker + (node.isOpened ? " s-open" : " s-close");
						if (node.childRoot) {
							node.childRoot.style.display = node.isOpened ? "" : "none";
						}
						record.treeNode.picker.style.display = "";
					}
					var padding = 0;
					if (parentNode) {
						node.isVisible = parentNode.isVisible && parentNode.isOpened;
						record.dataRow.style.display = node.isVisible ? "" : "none";
						record.dataRow.syraParentNode = parentNode.$uuid;
						if (!this.list.$item.$isNodeLazyLoad && !hasChildren && record.treeNode.icon) {
							padding = 18;
						}
					} else {
						node.isVisible = true;
					}
					record.treeNode.slot.style.paddingLeft = (record.treeNode.level * 18) + padding + "px";
					if (skinLevel) {
						record.dataRow.className += skinLevel;
						record.treeNode.slot.className += skinLevel;
						record.treeNode.item.className += skinLevel;
					}
				}
				if (hasChildren) {
					this.drawNodesOrder(node, node.children, level + 1);
				}
			}
		}
	},
	onEndFillList: function() {
		this._orderedNodes = [];
		for (var ii = 0, jj = this.tree.length; ii < jj; ii++) {
			var rootNode = this.tree[ii];
			rootNode.isOpened = true;
			this.drawNodesOrder(null, rootNode.children, 0);
		}
	},
	appendTreeviewCol: function(slot) {
		var col = this.list.cols.$treeview = this.list.appendCol();
		col.slot = slot;
		col.titleCell.className = this.cssNodeTitleCell;
		slot.appendChild(col.titleCell);
		if (this.$visibleFields.length == 0) {
			this.hasFixWith = false;
			col.groupHead.style.width = col.groupBody.style.width = col.widthValue = "100%";
		} else {
			this.hasFixWith = true;
			col.groupHead.style.width = col.groupBody.style.width = (col.widthValue = 100) + "px";
		}
		return col;
	},
	onRecordResize: function(record) {
		if (this.hasFixWith) {
			var slot = record.treeNode && record.treeNode.item;
			if (slot) {
				var index = record.getRecordIndex();
				if (index == 0) {
					this.onResizeMaxDiff = 0;
				}
				var diff = slot.scrollWidth - slot.clientWidth;
				if (diff > this.onResizeMaxDiff) {
					this.onResizeMaxDiff = diff;
				}
				if (this.onResizeMaxDiff && index == (this.list.records.length - 1)) {
					var col = this.list.cols.$treeview;
					col.groupHead.style.width = col.groupBody.style.width = (col.widthValue += this.onResizeMaxDiff) + "px";
					this.onResizeMaxDiff = 0;
				}
			}
		}
	},
	appendNode: function(record, slot) {
		var treeNode = record.treeNode = {
			slot: slot,
			item: document.createElement("div")
		};
		treeNode.item.className = this.cssNode;
		treeNode.picker = document.createElement("a");
		treeNode.picker.setAttribute("href", "#");
		treeNode.picker.setAttribute("data-s-picker", "list-tree-picker");
		if (this.list.$item.$isNodeLazyLoad) {
			treeNode.picker.className = this.cssNodePicker + " s-close";
			treeNode.picker.style.display = "";
		} else {
			treeNode.picker.className = this.cssNodePicker;
			treeNode.picker.style.display = "none";
		}
		treeNode.item.appendChild(treeNode.picker);
		if (this.list.selector.appendToRecord) {
			treeNode.selectorSlot = document.createElement("div");
			treeNode.selectorSlot.style.display = "none";
			treeNode.selectorSlot.className = this.cssNodeSelector;
			treeNode.item.appendChild(treeNode.selectorSlot);
		}
		var $bindings = this.list.$prototype.$treeview.$bindings;
		if ($bindings.$icon) {
			var name = record.dataset[$bindings.$icon];
			treeNode.icon = document.createElement("div");
			treeNode.icon.className = this.cssNodeIcon;
			treeNode.icon.style.backgroundImage = "url('" + document.site.$item.$iconPath + "tree/" + name + "')";
			treeNode.item.appendChild(treeNode.icon);
		}
		treeNode.desc = document.createElement("div");
		treeNode.desc.className = this.cssNodeDesc;
		treeNode.descriptionField = record.page.loadNewItem(treeNode.item.appendChild(treeNode.desc), {
			$bind: $bindings.$description,
			$isCellChild: true,
			$inplace: true
		}, record);
		treeNode.descriptionField.setState({
			$isHidden: false
		});
		treeNode.descriptionField.fieldValue.className += " " + this.cssNodeDescValue;
		treeNode.slot.appendChild(treeNode.item);
	},
	appendCellNode: function(record, slot) {
		var cell = document.createElement("td");
		cell.className = this.list.gridCss.cell + " " + this.cssNodeCell;
		if (this.list.$capability && this.list.$capability.reorder) {
			cell.className += " s-tree-record-reorder";
			document.site.ddManager.setDragSpot(cell, true);
		}
		slot.appendChild(cell);
		this.appendNode(record, cell);
		return cell;
	},
	appendCardNode: function(record, slot) {
		this.appendNode(record, slot);
		return slot;
	},
	onRemoveRecord: function() {
		/* previous code implementation
         if (this.parentNode && this.parentNode.children) {
         for (var ii = 0, jj = this.parentNode.children.length; ii < jj; ii++) {
         if (this.parentNode.children[ii] == this) {
         this.parentNode.children.splice(ii, 1);
         break;
         }
         }
         }
         if (this.children) {
         for (var ii = 0, jj = this.children.length; ii < jj; ii++) {
         delete this.children[ii].parentNode;
         }
         }
         if (this.nodes) {
         delete this.nodes[this.id];
         }
         if (this.record) {
         this.record.treeNode = null;
         }
         this.dispose();
         */
	},
	dispose: function() {
		this._uuidNodes = this._nodes = this.tree = this.list = this.$visibleFields = this._orderedNodes = null;
	}
});
