// syracuse-tablet/html/build/Gruntfile.js
"use strict";

module.exports = function(grunt) {
	// Project configuration.
	grunt.initConfig({
		pkg: grunt.file.readJSON('package.json'),
		copy: {

			/*
			 * copy bootstrap to metro-bootstrap to resolve dependencies quickly
			 */
			bootstrap: {
				expand: true,
				cwd: 'node_modules/bootstrap',
				src: '**/*.*',
				dest: 'node_modules/metro-bootstrap/app/bower_components/bootstrap',
			},

			/*
			 * Copy dependencies maintained in build folder to application folder
			 */
			deps_bootstrap_fonts: {
				expand: true,
				cwd: 'node_modules/bootstrap/dist/fonts',
				src: '**/*.*',
				dest: '../fonts',
			},
			deps_bootstrap_js: {
				expand: true,
				cwd: 'node_modules/bootstrap/dist',
				src: 'js/*.*',
				dest: '../deps/bootstrap',
			},
			deps_font_awesome: {
				expand: true,
				cwd: 'node_modules/font-awesome',
				src: ['css/*.*', 'fonts/*.*'],
				dest: '../deps/font-awesome',
			},


			js_precompile: {
				expand: true,
				cwd: '..',
				src: ['js/**/*.*', '!**/resources/**'],
				dest: 'tmp',
				/*
				options: {
					process: function (content, srcpath) {
						content = "define(function(require, exports, module) {\n" + content + "\n});";
				        return content;
				     }
				}
				*/
			}
		},

		clean: {
			/*
			 * remove copied bootstrap in metro-bootstrap
			 */
			bootstrap: [
				'node_modules/metro-bootstrap/app/bower_components'
			]
		},

		less: {
			development: {
				options: {
					compress: false,
					yuicompress: false,
					optimization: 2
				},
				files: {
					'../css/gen/windows.css': '../less/windows/main.less',
					'../css/gen/ios.css': '../less/ios/main.less',
					'../css/gen/android.css': '../less/android/main.less'
				}
			},
			release: {
				options: {
					compress: true,
					yuicompress: true,
					optimization: 2
				},
				files: {
					'../css/gen/windows.min.css': '../less/windows/main.less',
					'../css/gen/ios.min.css': '../less/ios/main.less',
					'../css/gen/android.min.css': '../less/android/main.less'
				}
			}

		},

		requirejs: {
			compile_min_debug: {
				options: {
					cjsTranslate: true,
					findNestedDependencies: true,
					baseUrl: '../../../',
					name: 'syracuse-tablet/html/js/main',
					out: '../js/main.joined.js',
					optimize: 'none'
				}
			},
			compile_min_release: {
				options: {
					cjsTranslate: true,
					findNestedDependencies: true,
					baseUrl: '../../../',
					name: 'syracuse-tablet/html/js/main',
					out: '../js/main.min.js'
				}
			}

		},


		concat: {
			options: {
				banner: '"use strict";\n',
				stripBanners: true
			},
			use_strict: {
				files: {
					'../js/main.joined.js': ['../js/main.joined.js'],
					'../js/main.min.js': ['../js/main.min.js']
				}
			},
		}
	});

	grunt.loadNpmTasks('grunt-contrib-less');
	grunt.loadNpmTasks('grunt-contrib-copy');
	grunt.loadNpmTasks('grunt-contrib-clean');
	grunt.loadNpmTasks('grunt-contrib-requirejs');
	grunt.loadNpmTasks('grunt-contrib-concat');

	grunt.registerTask('default', ['update_dev']);

	// Update 3rd party css, compile less
	grunt.registerTask('update_dev', ['copy:bootstrap',
		'less:development',
		'copy:deps_bootstrap_js',
		'copy:deps_bootstrap_fonts',
		'copy:deps_font_awesome',
		'clean:bootstrap'
	]);

	grunt.registerTask('release', [
		'copy:bootstrap',
		'less:development',
		'less:release',
		'copy:deps_bootstrap_js',
		'copy:deps_bootstrap_fonts',
		'copy:deps_font_awesome',
		'requirejs:compile_min_debug',
		'requirejs:compile_min_release',
		'concat:use_strict',
		'clean:bootstrap'
	]);
};
