// syracuse-ui/lib/page/dashboardPage.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var DesktopPage = require('./desktopPage').DesktopPage;
var drawHelper = require("syracuse-ui/lib/authoring/drawHelper");
function DashboardPage(){
}

exports.DashboardPage = helpers.defineClass(DashboardPage, DesktopPage, {
    dispose: function(){
        this._vignettes = this._vignettesMenus = null;
        DesktopPage.prototype.dispose.call(this);
    },
    registerVignette: function(vignette, $bind){
        this._vignettes[$bind || vignette.$item.$bind] = vignette;
    },
    onAuthoringAddNewField: function($newItem){
        if ($newItem.$fieldCategory == "link") {
            delete this._vignettesMenus.garbageLinks[$newItem.$fieldBind];
        }
    },
    onAuthoringExcludeItem: function(removedItem){
        if (removedItem.$bind && removedItem.isMenuItem) {
            this._vignettesMenus.garbageLinks[removedItem.$bind] = 1;
        }
    },
    saveGarbageLinks: function($article){
        $article.$garbageLinks = Object.keys(this._vignettesMenus.garbageLinks);
    },
    loadVignetteMenu: function(vignette, $vignette){
        var $field = this.$prototype.$properties[$vignette];
        if ($field) {
            vignette.$item.$vignette = $vignette;
            vignette.setTitle(vignette.$item.$title = $field.$title);
            if ($field.$links) {
                var $binds = Object.keys($field.$links);
                for (var ii = 0, jj = $binds.length; ii < jj; ii++) {
                    vignette.addMenuItem({
                        $bind: $binds[ii]
                    });
                }
            }
        }
    },
    newLoadVignetteMenu: function(vignette, $vignette){
        var $field = this.$prototype.$properties[$vignette];
        if ($field) {
            vignette.$item.$vignette = $vignette;
            this.registerVignette(vignette, $vignette);
            vignette.setTitle(vignette.$item.$title = $field.$title);
            vignette.layoutContent.$layout.$layoutType = "row";
            vignette.layoutContent.$layout.$widths = "33,33,33";
            vignette.layoutContent.ensureLayoutMode();
            var $items = [{
                $layoutType: "stack",
                $items: []
            }, {
                $layoutType: "stack",
                $items: []
            }, {
                $layoutType: "stack",
                $items: []
            }];
            this._vignettesMenus.$freeLinkBinds = Object.keys(this._vignettesMenus.freeLinks);
            var index = 0;
            for (var ii = 0, jj = this._vignettesMenus.$freeLinkBinds.length; ii < jj; ii++) {
                var $linkBind = this._vignettesMenus.$freeLinkBinds[ii];
                var $freeLink = this._vignettesMenus.freeLinks[$linkBind];
                if ($freeLink && $freeLink.$vignettes && $freeLink.$vignettes.indexOf($vignette) >= 0) {
                    $items[index++].$items.push({
                        $category: "link",
                        $skin: "s-mn-h2-link",
                        $bind: $linkBind
                    });
                    delete this._vignettesMenus.freeLinks[this._vignettesMenus.$freeLinkBinds[ii]];
                    if (index == 3) {
                        index = 0;
                    }
                }
            }
            for (var ii = 0, jj = $items.length; ii < jj; ii++) {
                vignette.layoutContent.createChildItem($items[ii], null, ii);
            }
            
        }
    },
    _addMenuToMenusBag: function($bind, $menu){
        if (!this._vignettesMenus.garbageLinks[$bind]) {
            if ($menu.$vignettes) {
                this._vignettesMenus.freeLinks[$bind] = $menu;
            }
            else {
                DesktopPage.prototype._addMenuToMenusBag.call(this, $bind, $menu);
            }
        }
        else {
            //debugger;
        }
    },
    _flagNewItems: function($newLinkBinds, $newVignetteBinds){
        if ($newLinkBinds.length > 0) {
            for (var ii = 0, jj = $newLinkBinds.length; ii < jj; ii++) {
                var menus = this.menuItems[$newLinkBinds[ii]];
                if (menus) {
                    for (var mm = 0, kk = menus.length; mm < kk; mm++) {
                        menus[mm].mn.className += " s-dash-new-link";
                    }
                }
            }
        }
        if ($newVignetteBinds.length > 0) {
            for (var ii = 0, jj = $newVignetteBinds.length; ii < jj; ii++) {
                var vignette = this._vignettes[$newVignetteBinds[ii]];
                if (vignette) {
                    var header = vignette._header;
                    if (vignette.isMenuGroup) {
                        if (vignette._header.style.display == "none" && vignette.layoutParent && vignette.layoutParent.boxParent.isSection) {
                            header = vignette.layoutParent.boxParent._header;
                        }
                    }
                    document.site.toggleClass(header, "s-dash-new-vignette", true);
                }
            }
        }
        
    },
    
    onMenuClick: function(menuItem){
        if (menuItem.$sourceBind) {
            var localize = this.ensurePageLocalize();
            switch (menuItem.$sourceBind) {
                case "$dash_apply_default":
                    var $article = helpers.object.clone(document.site.ensureDefaultArticle(this.$prototype.$article, this.$prototype), true);
                    var $newLinkBinds = this._vignettesMenus.$freeLinkBinds;
                    var $newVignetteBinds = this._vignettesMenus.$freeVignetteBinds;
                    this._vignettesMenus.freeLinks = {};
                    this.reloadLayout($article);
                    this.warnAdministrationUpdates();
                    this._flagNewItems($newLinkBinds, $newVignetteBinds);
                    this.showDiagnoses({
                        $diagnoses: [{
                            $message: localize.p_dashDefaultApplied,
                            $severity: "info",
                            $links: {
                                "$dash_save_authoring": {
                                    $title: localize.p_dashSaveAuthoring
                                },
                                "$dash_open_authoring": {
                                    $title: localize.p_dashOpenAuthoring
                                }
                            }
                        }]
                    });
                    return false;
                case "$dash_add_new_items":
                    this.layoutContent.ensureContentIsLoaded();
                    var $newLinkBinds = this._vignettesMenus.$freeLinkBinds;
                    var $newVignetteBinds = this._vignettesMenus.$freeVignetteBinds;
                    if ($newVignetteBinds.length > 0) {
                        var $field;
                        var layout = this.layoutContent;
                        for (var ii = 0, jj = $newVignetteBinds.length; ii < jj; ii++) {
                            $field = this.$prototype.$properties[$newVignetteBinds[ii]];
                            var $item = {};
                            if ($field.$format == "$menu") {
                                drawHelper.onAddNewVignette({
                                    awArticle: this
                                }, this.layoutContent, {
                                    $category: "menus",
                                    $title: $field.$title,
                                    $vignette: $newVignetteBinds[ii]
                                }, "top");
                            }
                            else {
                                drawHelper.onAddNewField({
                                    awArticle: this
                                }, this.layoutContent, {
                                    $fieldBind: $newVignetteBinds[ii],
                                    $fieldCategory: "field"
                                }, "top");
                            }
                        }
                    }
                    this._vignettesMenus.$freeLinkBinds = Object.keys(this._vignettesMenus.freeLinks);
                    if (this._vignettesMenus.$freeLinkBinds.length > 0) {
                        for (var ii = 0, jj = this._vignettesMenus.$freeLinkBinds.length; ii < jj; ii++) {
                            var $linkBind = this._vignettesMenus.$freeLinkBinds[ii];
                            var $freeLink = this._vignettesMenus.freeLinks[$linkBind];
                            if ($freeLink && $freeLink.$vignettes) {
                                var vignette;
                                for (var mm = 0, kk = $freeLink.$vignettes.length; !vignette && mm < kk; mm++) {
                                    var sibling = this.findSiblingMenu($freeLink.$vignettes[mm]);
                                    if (sibling) {
                                        drawHelper.onAddNewField({
                                            awArticle: this
                                        }, sibling, {
                                            $fieldBind:$linkBind,
                                            $fieldCategory: "link",
											$fieldSkin: "s-mn-h2-link",
											$category :"link"	
                                        }, "top");
                                        delete this._vignettesMenus.freeLinks[$linkBind];
                                    }
                                }
                            }
                        }
                    }
                    this.layoutValidator.validate(this.layoutContent, true);
                    
                    this._vignettesMenus.freeLinks = {};
                    this.warnAdministrationUpdates();
                    this._flagNewItems($newLinkBinds, $newVignetteBinds);
                    this.showDiagnoses({
                        $diagnoses: null
                    });
                    this.showDiagnoses({
                        $diagnoses: [{
                            $message: localize.p_dashDefaultApplied,
                            $severity: "info",
                            $links: {
                                "$dash_save_authoring": {
                                    $title: localize.p_dashSaveAuthoring
                                },
                                "$dash_open_authoring": {
                                    $title: localize.p_dashOpenAuthoring
                                }
                            }
                        }]
                    });
                    return false;
                case "$dash_open_authoring":
                    this.showDiagnoses({
                        $diagnoses: null
                    });
                    document.site.openAuthoringPage(true);
                    return false;
                case "$dash_save_authoring":
                    return false;
            }
        }
        return true;
    },
    getLoadedVignetteMenus: function(){
        this._vignettesMenus.loaded = {};
        var $binds = Object.keys(this.menuItems);
        for (var ii = 0, jj = $binds.length; ii < jj; ii++) {
            var menus = this.menuItems[$binds[ii]];
            if (menus.length > 0) {
                var $vignettes = menus[0].$vignettes;
                if ($vignettes) {
                    for (var mm = 0, kk = $vignettes.length; mm < kk; mm++) {
                        this._vignettesMenus.loaded[$vignettes[mm]] = this.$prototype.$properties[$vignettes[mm]];
                    }
                }
            }
        }
    },
    findSiblingMenu: function($vignette){
        var vignette = this._vignettes[$vignette];
        if (!vignette) {
            var $binds = Object.keys(this.menuItems);
            for (var ii = 0, jj = $binds.length; ii < jj; ii++) {
                var menus = this.menuItems[$binds[ii]];
                if (menus.length > 0 && menus[0].$vignettes) {
                    if (menus[0].$vignettes.indexOf($vignette) >= 0) {
                        return menus[0];
                    }
                }
            }
        }
        return null;
    },
    warnAdministrationUpdates: function(){
        this._vignettesMenus.$freeLinkBinds = Object.keys(this._vignettesMenus.freeLinks);
        var $vignettesBinds = Object.keys(this.$prototype.$properties);
        this._vignettesMenus.$freeVignetteBinds = [];
        if (this._vignettesMenus.$freeLinkBinds.length > 0) {
            this.getLoadedVignetteMenus();
        }
        var $definedBinds = {};
        this.getDefinedFieldsObject(this.$item.$layout.$items, $definedBinds);
        for (var ii = 0, jj = $vignettesBinds.length; ii < jj; ii++) {
            var $bind = $vignettesBinds[ii];
            if (!this._vignettes[$bind]) {
                var $field = this.$prototype.$properties[$bind];
                if ($field) {
                    if ($field.$format == "$menu") {
                        if (this._vignettesMenus.$freeLinkBinds.length > 0) {
                            for (var mm = 0, kk = this._vignettesMenus.$freeLinkBinds.length; mm < kk; mm++) {
                                var $freeLink = this._vignettesMenus.freeLinks[this._vignettesMenus.$freeLinkBinds[mm]];
                                if ($freeLink.$vignettes && $freeLink.$vignettes.indexOf($bind) >= 0) {
                                    //vignette has new menu Item 
                                    if (!this._vignettesMenus.loaded[$bind]) {
                                        this._vignettesMenus.$freeVignetteBinds.push($bind);
                                    }
                                    break;
                                }
                            }
                        }
                    }
                    else {
                        if ($bind != "$pageViewSelector" && !$definedBinds[$bind]) {
                            this._vignettesMenus.$freeVignetteBinds.push($bind);
                        }
                    }
                }
            }
        }
        if (this._vignettesMenus.$freeLinkBinds.length > 0 || this._vignettesMenus.$freeVignetteBinds.length > 0) {
            var $message;
            var localize = this.ensurePageLocalize();
            if (this._vignettesMenus.$freeLinkBinds.length > 0) {
                if (this._vignettesMenus.$freeVignetteBinds.length > 0) {
                    $message = localize.p_dashNewLinkVignette.replace("{0}", this._vignettesMenus.$freeLinkBinds.length);
                    $message = $message.replace("{1}", this._vignettesMenus.$freeVignetteBinds.length);
                }
                else {
                    $message = localize.p_dashNewLink.replace("{0}", this._vignettesMenus.$freeLinkBinds.length);
                }
            }
            else {
                $message = localize.p_dashNewVignette.replace("{0}", this._vignettesMenus.$freeVignetteBinds.length);
            }
            this.showDiagnoses({
                $diagnoses: [{
                    $message: $message,
                    $severity: "warning",
                    $links: {
                        "$dash_apply_default": {
                            $title: localize.p_dashNewApplyDefault
                        },
                        "$dash_add_new_items": {
                            $title: localize.p_dashNewBestPlace
                        },
                        "$dash_open_authoring": {
                            $title: localize.p_dashOpenAuthoring
                        }
                    }
                }]
            });
        }
    },
    loadBox: function(initData, $initDiagnoses){
        this._vignettesMenus = {
            freeLinks: {},
            garbageLinks: {},
            loaded: {}
        };
        if (this.$item.$garbageLinks) {
            for (var ii = 0, jj = this.$item.$garbageLinks.length; ii < jj; ii++) {
                this._vignettesMenus.garbageLinks[this.$item.$garbageLinks[ii]] = 1;
            }
        }
        this._vignettes = {};
        this.isDashBoard = true;
        this.ensurePageLocalize();
        this.$skin = this.$item.$skin || "s-dash";
        this.$defaultSkinBlock = "s-dash-h2";
        this.$defaultSkinSection = "s-dash-h1";
        this._defaultTitle = this.pageLocalize.p_dashboardTitle;
        this.$isSearchFacetEnabled = false;
        this.$autoFetch = false;
        DesktopPage.prototype.loadBox.call(this, initData, $initDiagnoses);
        this.warnAdministrationUpdates();
    }
});
