// syracuse-ui/lib/page/menuBar.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function MenuBar(){
}

exports.MenuBar = helpers.defineClass(MenuBar, null, {
    load: function(page){
        this.preferences = document.site.ensurePreferences();
        this.preferences.menuBar = this.preferences.menuBar || {};
        
        page.initializeNewItem(this);
        if (this.preferences.menuBar.isDocked === undefined) {
            this.preferences.menuBar.isDocked = true;
        }
        this._slot = document.createElement("div");
        this._slot.className = this.page.$skin + "-menubar-slot";
        this._slot.style.display = "none";
        this._slot.width = "180px";
        this.$$slot = $(this.page._item.appendChild(this._slot));
        
        this._bar = document.createElement("div");
        this._bar.className = "s-page-menu-bar";
        this._slot.appendChild(this._bar);
        //    this._appendHeader();
        this.pickerBar = document.createElement("a");
        this.pickerBar.setAttribute("data-s-picker", "s-bar-collapse");
        this.pickerBar.className = "s-page-menu-bar-picker";
        this._bar.appendChild(this.pickerBar);
        
        this._body = document.createElement("div");
        this._body.className = "s-page-menu-bar-body";
        this._$$body = $(this._body);
        this._bar.appendChild(this._body);
        
        this.menuGroups = this.page.appendArticleMenus(this._body);
        
        this._bindEvents(true);
        this.ensureState();
    },
    hideBar: function(){
        this.isHidden = true;
        if (this._slot) {
            this._slot.style.display = "none";
        }
    },
    onMenuDataFilled: function(){
        if (!this.isSlotVisible && !this.page.authoringPage && !this.isHidden) {
            this._slot.style.display = "";
            this.isSlotVisible = true;
            if (this.page.layoutSlot.style.display != "none") {
                document.site.resize();
            }
        }
    },
    _appendHeader: function(){
        this._header = document.createElement("header");
        this._header.className = "s-page-menu-bar-header";
        
        var collapse = document.createElement("a");
        collapse.setAttribute("data-s-picker", "s-bar-collapse");
        collapse.className = "s-page-menu-bar-collapse";
        this._header.appendChild(collapse);
        this._$$header = $(this._header);
        
        this.dockMode = document.createElement("a");
        this.dockMode.setAttribute("data-s-picker", "s-bar-dockMode");
        this.dockMode.className = "s-page-menu-bar-docked";
        this.dockMode.style.display = "none";
        
        this._header.appendChild(this.dockMode);
        this._bar.appendChild(this._header);
        
    },
    onClickPicker: function(pickerId){
        switch (pickerId) {
            case "s-bar-dockMode":
                this.preferences.menuBar.isDocked = !this.preferences.menuBar.isDocked;
                break;
            case "s-bar-collapse":
                this.preferences.menuBar.isCollapsed = !this.preferences.menuBar.isCollapsed;
                break;
        }
        this.ensureState();
    },
    _bindEvents: function($bind){
        var self = this;
        if ($bind) {
            self.$$slot.bind("mouseenter mouseleave", function(event){
                clearTimeout(self._closeTimeout);
                if (event.type == "mouseleave" && !self.preferences.menuBar.isCollapsed && !self.preferences.menuBar.isDocked) {
                    self._closeTimeout = setTimeout(function(){
                        self.preferences.menuBar.isCollapsed = !self.preferences.menuBar.isCollapsed;
                        self.ensureState();
                    }, 300);
                }
            });
        }
        else {
            if (self.$$slot) {
                self.$$slot.unbind();
            }
        }
    },
    ensureState: function(){
        if (this.preferences.menuBar.isCollapsed) {
            this._slot.style.width = "15px";
            //this._header.style.display = "none";
            this._body.style.display = "none";
            this.pickerBar.style.width = "15px";
            this.pickerBar.className = "s-page-menu-bar-picker s-close";
            if (!this.preferences.menuBar.isDocked) {
                this._bar.style.width = "15px";
            }
        }
        else {
            this.pickerBar.style.width = "";
            this.pickerBar.className = "s-page-menu-bar-picker";
            if (this.preferences.menuBar.isDocked) {
                //this.dockMode.className = "s-page-menu-bar-docked";
                this._bar.style.width = "";
                this._slot.style.width = "180px";
                this._bar.style.position = "";
            }
            else {
                //this.dockMode.className = "s-page-menu-bar-float";
                this._slot.style.width = "15px";
                document.site.setZIndex(this._bar);
                this._bar.style.position = "absolute";
                this._bar.style.width = "180px";
                this._bar.style.top = "0px";
                this._bar.style.right = "0px";
            }
            //this._header.style.display = "";
            this._body.style.display = "";
            if (this.isSlotVisible) {
                this._slot.style.display = "";
            }
        }
    },
    onWindowResize: function(){
        if (this._bar && this._slot) {
            var container;
            /*if (this.page.$dialogMode) {
             this.dockMode.style.display = "none";
             }*/
            var height = $(this.page.scrollview).height();
            if (height) {
                this._bar.style.height = height + "px";
                this.pickerBar.style.height = height + "px";
                //this._body.style.height = (height - this._$$header.outerHeight(true)) + "px";
                this._body.style.height = height + "px";
            }
        }
    },
    toggleItemAuthoring: function(enable){
        if (enable) {
        }
        else {
            if (this.menuGroups) {
                document.site.authorPage.toggleItemAuthoring(this.menuGroups, enable);
            }
        }
    },
    dispose: function(){
        this._bindEvents(false);
        this.menuGroups = this.$$slot = this._slot = this._bar = this._header = null;
        this._$$body = this._body = this.dockMode = this.page = null;
    }
});
