// syracuse-ui/lib/page/menuBar.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function MenuBar() {}

exports.MenuBar = helpers.defineClass(MenuBar, null, {
	load: function(page) {
		page.initializeNewItem(this);
		this._slot = document.createElement("div");
		this._slot.className = this.page.$skin + "-menubar-slot";
		this._slot.style.display = "none";
		this._slot.width = "200px";
		this.preferences = document.site.ensurePreferences();
		if (page.$pageCategory == "vignette" || page.$pageCategory == "dashboard") {
			this.page._item.insertBefore(this._slot, this.page._item.firstChild);
			if (page.$pageCategory == "vignette") {
				this.preferences = {}; //temp for preference management
			}
		} else {
			this.page._item.appendChild(this._slot);
		}
		if (!this.preferences.menuBar) {
			this.preferences.menuBar = {
				isDocked: true,
				isCollapsed: false
			};
		}
		this.$$slot = $(this._slot);

		this._bar = document.createElement("div");
		this._bar.className = this.page.$skin + "-menubar";
		this.$$bar = $(this._bar);
		this._slot.appendChild(this._bar);

		this.appendPickerBar();
		this._body = document.createElement("div");
		this._body.className = this.page.$skin + "-menubar-body";
		this._$$body = $(this._body);
		this._bar.appendChild(this._body);
		this.page.appendArticleMenus(this._body);
		this.ensureState();
	},
	appendPickerBar: function() {
		var self = this;
		self.resizeBar = document.createElement("a");
		self.resizeBar.className = this.page.$skin + "-menubar-resizer";
		self.openerPicker = document.createElement("a");
		self.openerPicker.className = this.page.$skin + "-menubar-opener";
		self.openerPicker.setAttribute("data-s-picker", "s-bar-collapse");
		self.openerPicker.style.display = "none";
		self.resizeBar.appendChild(self.openerPicker);
		self._bar.appendChild(self.resizeBar);
		self._resizer = document.site.setResizable({
			source: self,
			slot: self._slot,
			dragSpot: null,
			direction: {
				left: true
			},
			resizerSpot: self.resizeBar,
			minWidth: 200,
			onResize: function(resizer, moving) {
				if (!moving) {
					document.site.toggleClass(self.resizeBar, "s-resizing", document.site.resizing);
					document.site.toggleClass(self.openerPicker, "s-resizing", document.site.resizing);
				}
			}
		});
		self._resizer.isEnabled = !self.preferences.menuBar.isCollapsed;
	},
	toggleBar: function(show) {
		this.isHidden = !show;
		if (this._slot) {
			this._slot.style.display = this.isHidden ? "none" : "";
		}
	},
	onMenuDataFilled: function(menuItem) {
		if (!this.isSlotVisible && !this.isHidden) {
			if (menuItem.boxParent.menuGroupRoot) {
				if (menuItem.boxParent.menuGroupRoot.$item.$isMenuPage) {
					this.isSlotVisible = true;
					if (!this.page.authoringPage) {
						this._slot.style.display = "";
						if (this.page.layoutSlot.style.display != "none") {
							document.site.resize();
						}
					}
				}
			}
		}
	},
	collapse: function() {
		this._resizer.isEnabled = !(this.preferences.menuBar.isCollapsed = !this.preferences.menuBar.isCollapsed);
		this.ensureState();
		document.site.resize();
	},
	onClickPicker: function(btn) {
		switch (btn.getAttribute("data-s-picker")) {
			case "s-bar-dockMode":
				this.preferences.menuBar.isDocked = !this.preferences.menuBar.isDocked;
				this.ensureState();
				document.site.resize();
				break;
			case "s-bar-collapse":
				this.collapse();
				break;
		}
	},
	ensureState: function() {
		if (!this.isHidden && !this._onAuthoring) {
			if (this.preferences.menuBar.isCollapsed) {
				this.isCollapsed = true;
				this.openedWidth = this.$$slot.outerWidth();
				this._slot.style.width = "10px";
				this._body.style.display = "none";
				document.site.toggleClass(this.resizeBar, "s-close", true);
				document.site.toggleClass(this.openerPicker, "s-close", true);
				if (!this.preferences.menuBar.isDocked) {
					this._bar.style.width = "10px";
				}
			} else {
				this.isCollapsed = false;
				this.resizeBar.style.width = "";
				document.site.toggleClass(this.resizeBar, "s-close", false);
				document.site.toggleClass(this.openerPicker, "s-close", false);
				if (this.preferences.menuBar.isDocked) {
					this._bar.style.width = "";
					this._slot.style.width = (this.openedWidth || 200) + "px";
					this._bar.style.position = "";
				} else {
					this._slot.style.width = "10px";
					document.site.setZIndex(this._bar);
					this._bar.style.position = "absolute";
					this._bar.style.width = "200px";
					this._bar.style.top = "0px";
					this._bar.style.right = "0px";
				}
				this._body.style.display = "";
				if (this.isSlotVisible) {
					this._slot.style.display = "";
				}
			}
			this._resizeWidthPage();
		}
	},
	onWindowResize: function() {
		if (this._bar && this._slot) {
			var $$scrollview = $(this.page.scrollview);
			var height = $$scrollview.height();
			if (height) {
				if (this.preferences.menuBar.isCollapsed != this.isCollapsed) {
					this.ensureState();
				}
				height += $$scrollview.position().top;
				this._bar.style.height = height + "px";
				this.resizeBar.style.height = height + "px";
				this.openerPicker.style.display = "";
				this.openerPicker.style.top = ((height - this.openerPicker.clientHeight) / 2) + "px";
				this._body.style.height = height + "px";
				if (this.authoringOverlay) {
					var style = this.authoringOverlay.style;
					var offset = this.$$bar.offset();
					style.top = offset.top + "px";
					style.left = offset.left + "px";
					style.height = this.$$bar.outerHeight(true) + "px";
					style.width = this.$$bar.outerWidth(true) + "px";
				}
				this._resizeWidthPage();
			}
		}
	},
	_resizeWidthPage: function() {
		if (this.page.dialogWrapper && !this._onAuthoring) {
			var width = this.page.dialogWrapper.$$dialogSlot.width();
			if (width) {
				this.page.scrollview.style.width = (width - this.$$bar.outerWidth(true)) + "px";
			}
		}
	},
	toggleItemAuthoring: function(enable, isOnPreviewMode) {
		if (enable) {
			if (!this.authoringOverlay) {
				this.authoringOverlay = document.createElement("div");
				this.authoringOverlay.className = "s-fusion-bar-overlay";
				document.site.setZIndex(this.authoringOverlay);
				document.site.layoutSlot.appendChild(this.authoringOverlay);
			}
			this._onAuthoring = true;
			this._slot.style.display = this.authoringOverlay.style.display = (isOnPreviewMode && (!this.isHidden && this.isSlotVisible)) ? "" : "none";
			this.onWindowResize();
		} else {
			if (this.authoringOverlay) {
				document.site.removeDomChild(this.authoringOverlay);
				delete this.authoringOverlay;
			}
			this._onAuthoring = false;
			this.ensureState();
		}
	},
	dispose: function() {
		if (this._resizer) {
			this._resizer.dispose();
			this._resizer = null;
		}
		if (this.authoringOverlay) {
			document.site.removeDomChild(this.authoringOverlay);
			delete this.authoringOverlay;
		}
		this.$$slot = this._slot = this._bar = this.$$bar = null;
		this._$$body = this._body = this.page = null;
	}
});
