// syracuse-ui/lib/page/navigationPage.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
//var LandingPageDesigner = require("syracuse-ui/lib/authoring/landingPage/landingPageDesigner").LandingPageDesigner;
var DesktopPage = require('./desktopPage').DesktopPage;
var ModuleList = require('./navigation/moduleList').ModuleList;
var ModuleBar = require('./navigation/moduleBar').ModuleBar;

function NavigationPage() {}

exports.NavigationPage = helpers.defineClass(NavigationPage, DesktopPage, {
	designItem: function(open) {
		//      document.site.designItem(open, this, LandingPageDesigner);
	},
	appendHeader: function() {
		DesktopPage.prototype.appendHeader.call(this);
		this.designerOpenerLink = document.createElement("a");
		this.designerOpenerLink.className = "s-navpage-designer-picker";
		this.designerOpenerLink.setAttribute("data-s-picker", "openDesigner");
		this.designerOpenerLink.title = this.localize.landingPageOpenDesigner;
		this.designerOpenerLink.style.display = "none";
		this.headerCoreRight.appendChild(this.designerOpenerLink);
	},
	onClickPicker: function(picker, event) {
		if (picker == this.designerOpenerLink) {
			this.designer.toggleDesigner();
			event.stopPropagation();
			return false;
		}
		return true;
	},
	ensureDefaultArticle: function($article, $prototype) {
		return {
			$layout: {
				$items: []
			}
		};
	},
	createField: function($field, $item, boxParent, $class) {
		switch ($item.$bind) {
			case "modules":
				return new ModuleList();
		}
		return null;
	},
	loadBox: function(initData, $initDiagnoses) {
		this.isNavigationPage = true;
		this.$skin = this.$item.$skin || "s-navpage";
		this.$defaultSkinBlock = "s-navpage-h2";
		this.$defaultSkinSection = "s-navpage-h1";
		this._defaultTitle = this.localize.dskpNavigationPageTitle;
		this.$autoFetch = false;
		var data = {};
		if (initData) {
			var $keys = Object.keys(initData);
			for (var ii = 0, jj = $keys.length; ii < jj; ii++) {
				if ($keys[ii] != "$prototype" && $keys[ii] != "$article") {
					data[$keys[ii]] = initData[$keys[ii]];
				}
			}
		}
		DesktopPage.prototype.loadBox.call(this, data, $initDiagnoses);
		if (!this.moduleBar) {
			(this.moduleBar = new ModuleBar()).load(this);
		}
	},
	_ensureMenus: function() {},
	dispose: function() {
		if (this.moduleBar) {
			this.moduleBar.dispose();
			this.moduleBar = null;
		}
		DesktopPage.prototype.dispose.call(this);
	}

});
