// syracuse-ui/lib/page/pageBuilder.js
"use strict";
var _helpers = require('syracuse-core/lib/helpers');
var _localizationConverter = require('syracuse-ui/lib/localization/localizationConverter');
var DesktopPage = require('syracuse-ui/lib/page/desktopPage').DesktopPage;
var UserPreferences = require('syracuse-ui/lib/page/tools/userPreferences').UserPreferences;
var _childPages = [];

var _queryFacets = ["$query", "$bulk", "$lookup", "$select", "$cube"];
var _autoFetchFacets = _queryFacets.concat(["$details", "$edit", "$summary"]);

exports.getLoaded = function() {
	return _childPages;
};

exports.initialize = function(page) {
	if (!page.isSiteRegisterDisabled) {
		if (_childPages.indexOf(page) < 0) {
			_childPages.push(page);
		}
	}
	page.page = page;

	page._childItemOffset = page._childItemOffset || 1; //initialize for site
	if (!page.id) {
		page.id = syra_site._childPageOffset++;
		syra_item.register(page);
	}
};

exports.resize = function() {
	for (var ii = 0, jj = _childPages.length; ii < jj; ii++) {
		_childPages[ii].resizeArticle(true);
	}
};

function _endLoad(page, $itemPage, $representation, onLoaded) {
	var $protoArticle = page.$prototype.$article;
	var $article = page.initializePage ? page.initializePage($itemPage) : $representation.$article;
	if ($article) {
		delete $article.$menus;
	} else {
		$article = $protoArticle;
	}
	$article = exports.ensureDefaultArticle(page, $article, page.$prototype);
	if (page.$prototype.$menus) {
		$article.$menus = page.$prototype.$menus;
	} else {
		if ($protoArticle && $protoArticle.$menus) {
			$article.$menus = $protoArticle.$menus;
		}
	}
	/*if (page.$prototype && page.$prototype.$localization && page.$prototype.$hackLocalization) {
     _localizationConverter.applyHackLocalization(page.$prototype, $article);
     }*/
	if (page.$prototype && page.$prototype.$localization && page.$prototype.$hackBlockRank) {
		if (!$article.$isBlockRankSet) {
			$article.$isBlockRankSet = true;
			if ($article != $protoArticle) {
				if ($protoArticle) {
					$protoArticle.$isBlockRankSet = true;
				}
				_localizationConverter.applyHackBlockRank($article, page.$prototype.$hackBlockRank);
			}
		}
	}
	if (page.$prototype && page.$prototype.$properties) {
		if (page.$prototype.$properties.contract) {
			page.$prototype.$properties.contract.$isHidden = true;
		}
	}

	/*if (page.$prototype && page.$prototype.$links && page.$facet == "$details") {
     page.$prototype.$links.LK_EMB = {
     $target: "embedded",
     $title: "Test embedded",
     $type: "application/json;vnd.sage=syracuse",
     $url: "{$baseUrl}/endPoints?representation=endPoint.$query",
     };
     }*/
	if ($itemPage.inlinePageHost) {
		page.$isTitleHidden = true;
		page.inlinePageHost = $itemPage.inlinePageHost;
		delete $itemPage.inlinePageHost;
		page.inlinePageHost.onPreInitializeInlinePage(page, $article);
	}

	var layoutSlotStyle = $itemPage.layoutSlot && $itemPage.layoutSlot.style && $itemPage.layoutSlot.style.display;

	syra_site.initializeNewItem(page, $article, $itemPage.boxParent);
	syra_localizer.applyPageLocalization(page);

	page.$autoFetch = $itemPage.$autoFetch;
	if (page.$autoFetch === undefined) {
		if (page.$facet == "$edit" && $itemPage.initData) {
			page.$autoFetch = false;
		} else {
			page.$autoFetch = _autoFetchFacets.indexOf(page.$facet) >= 0;
		}
	}

	page.$item.externalAdapter = $itemPage.externalAdapter;
	page.$views = $representation.$views;
	page.loadBox($itemPage.initData);
	page.isPageLoaded = true;
	if ($representation.$diagnoses && $representation != $itemPage.initData) {
		syra_diagnose.showDiagnoses({
			$diagnoses: $representation.$diagnoses
		}, page);
	} else {
		page.diagnosesPanel && page.diagnosesPanel.refresh();
	}
	$itemPage.$pageCategoryClass = $itemPage.boxParent = $itemPage.layoutSlot = $representation = $itemPage.dialogWrapper = null;
	if (layoutSlotStyle !== undefined) {
		page.layoutSlot.style.display = layoutSlotStyle;
	}
	if ($itemPage.onMainPagechange) {
		delete $itemPage.onMainPagechange;
		syra_site.refreshMainPageSecurity(page);
		syra_dom.hide(syra_site.body, false);
	}
	syra_site.ensureArticleVisibility(page);
	if (page.$autoFetch) {
		page.fetch({
			isPageLoading: true
		}, function(data, response, requestUrl) {
			if (page && !page.disposed) {
				if (!data) {
					data = {
						$diagnoses: [{
							severity: "error",
							message: syra_local.flFetchDataFailed
						}]
					};
					var $binds = Object.keys(page.boundFields);
					for (var ii = 0, jj = $binds.length; ii < jj; ii++) {
						var $bind = [$binds[ii]];
						var field = page.boundFields[$bind];
						field = field && field[0];
						if (field && field.isArrayField) {
							data[$bind] = [];
						}
					}
				}
				page.startChange();
				page.applyChange(data, response, requestUrl);
				page.endChange();
				page.onAfterLoad();

			}
		});
	} else {
		page.onAfterLoad();
	}
	onLoaded(page);
}

exports.load = function($itemPage, onLoaded) {
	var $representation = $itemPage.$representation;
	var $pageCategory = ($representation.$article ? $representation.$article.$category : null) || $itemPage.$category;
	$itemPage.openerUrlSegments = $itemPage.openerUrlSegments || {};
	if (!$pageCategory) {
		switch ($itemPage.openerUrlSegments.representationRoot) {
			case "landingVignetteSelect":
				$pageCategory = "landingVignetteSelect";
				$itemPage.$autoFetch = true;
				break;
			default:
				switch ($itemPage.openerUrlSegments.$facet) {
					case "$search":
					case "$dashboard":
						$pageCategory = $itemPage.openerUrlSegments.$facet.slice(1);
						break;
					case "$navigation_edit":
					case "$navigation":
						$pageCategory = "navigation";
						$itemPage.initData = $representation;
						break;
					case "$landing_edit":
					case "$landing":
						$pageCategory = "landing";
						$itemPage.initData = $representation;
						break;
					default:
						$pageCategory = "page";
						break;
				}
				break;
		}
	}
	var page = new($itemPage.$pageCategoryClass || syra_widgets.pageCategories[$pageCategory] || DesktopPage)();
	var $userPreferencesLink = $representation.$links && $representation.$links.$userPreferences;
	if ($userPreferencesLink) {
		(page.userPreferences = new UserPreferences()).load(page, $userPreferencesLink, $representation.userPreferences && $representation.userPreferences.content);
	}
	page.id = syra_site._childPageOffset++;
	var openerUrlSegments = page.openerUrlSegments = $itemPage.openerUrlSegments;
	page.$pageCategory = $pageCategory;
	page.dialogWrapper = $itemPage.dialogWrapper;
	page.$pageCollaborationUrl = $representation.$pageCollaborationUrl;
	page.$authorUrl = $representation.$authorUrl;
	page.$facet = $itemPage.$facet || openerUrlSegments.$facet;
	page.$breadCrumb = $representation.$links && $representation.$links.$breadcrumb;

	page.$isEditMode = $itemPage.$isEditMode || page.$facet == "$edit";
	page.$isEditMode && syra_workingCopy.add(page);


	if ($itemPage.vignetteField) {
		page.vignetteField = $itemPage.vignetteField;
		delete $itemPage.vignetteField;
	}
	var hackFusion = page.$facet.split("_");
	if (hackFusion.length > 1 && hackFusion[0] == "$fusion") {
		page.$isEditMode = true;
		page.$facet = hackFusion[1];
		page.$fusionPageMeta = $itemPage.$fusionPageMeta;
		$itemPage.$autoFetch = false;
	}
	page.layoutSlot = $itemPage.layoutSlot;
	page.$prototype = $representation.$prototype;
	var view = openerUrlSegments && openerUrlSegments.params && openerUrlSegments.params.view;
	if (view == "calendar") {
		page.$prototype.$isCalendarView = true;
	}
	page.page = page;

	syra_widgets.ensureMediaTypes(page.$prototype, function() {
		_endLoad(page, $itemPage, $representation, onLoaded);
	});
};

exports.activateQueryList = function(page) {
	if (!page.isFusionPage && !page.isVignettePage && _queryFacets.indexOf(page.$facet) >= 0) {
		var list = syra_item.findList(page);
		if (list) {
			page.setShortcutSpyerList(list);
			var filterRecord = list.builder && list.builder.filterRecord;
			filterRecord && filterRecord.setFocus();
		}
	}
};

exports.disposePage = function(page) {
	syra_dlg.closeAll(page);
	page.userPreferences && page.userPreferences.dispose();
	if (!page.isVignettePage) {
		var found = _childPages.indexOf(page);
		if (found >= 0) {
			_childPages.splice(found, 1);
		}
	}
	page.userPreferences = null;
};


exports.autoScroll = function(scrollViewRect, event) {
	var step = 40,
		margin = 80;
	if (event.pageY > (scrollViewRect.bottom - margin)) {
		var scrollTop = scrollViewRect.item.scrollTop;
		if (scrollTop < (scrollViewRect.scrollHeight - scrollViewRect.height)) {
			scrollViewRect.item.scrollTop = Math.min(scrollTop + step, scrollViewRect.scrollHeight - scrollViewRect.height);
		}
	} else {
		if (event.pageY < (scrollViewRect.top + margin)) {
			var scrollTop = scrollViewRect.item.scrollTop;
			scrollViewRect.item.scrollTop = Math.max(scrollTop - step, 0);
		}
	}
	if (event.pageX > (scrollViewRect.right - margin)) {
		var scrollLeft = scrollViewRect.item.scrollLeft;
		if (scrollLeft < (scrollViewRect.scrollWidth - scrollViewRect.width)) {
			scrollViewRect.item.scrollLeft = Math.min(scrollLeft + step, scrollViewRect.scrollWidth - scrollViewRect.width);
		}
	} else {
		if (event.pageX < (scrollViewRect.left + margin)) {
			var scrollLeft = scrollViewRect.item.scrollLeft;
			scrollViewRect.item.scrollLeft = Math.max(scrollLeft - step, 0);
		}
	}
};

exports.getDefinedFieldBinds = function($items, $binds) {
	$binds = $binds || [];
	if ($items) {
		for (var ii = 0, jj = $items.length; ii < jj; ii++) {
			var $item = $items[ii];
			if ($item.$bind) {
				$binds.push($item.$bind);
			} else {
				if ($item.$items) {
					$binds = exports.getDefinedFieldBinds($item.$items, $binds);
				}
				if ($item.$layout && $item.$layout.$items) {
					$binds = exports.getDefinedFieldBinds($item.$layout.$items, $binds);
				}
			}
		}
	}
	return $binds;
};


exports.getDefinedFieldList = function($items, $list) {
	if (!$list) {
		$list = [];
	}
	if ($items) {
		for (var ii = 0, jj = $items.length; ii < jj; ii++) {
			var $item = $items[ii];
			if ($item.$bind) {
				$list.push($item);
			} else {
				if ($item.$items) {
					$list = exports.getDefinedFieldList($item.$items, $list);
				}
				if ($item.$layout && $item.$layout.$items) {
					$list = exports.getDefinedFieldList($item.$layout.$items, $list);
				}
			}
		}
	}
	return $list;
};


function _buildQueryArticle(page, $article, $prototype, isCube) {
	var $items = syra_layoutUpdater.getItemSettings(page, $article.$layout.$items, true) || [];
	var $defined = {};
	for (var ii = 0, jj = $items.length; ii < jj; ii++) {
		if ($items[ii].$bind) {
			$defined[$items[ii].$bind] = $items[ii];
		}
	}

	var $binds = Object.keys($prototype.$properties);
	var $field, $array = [],
		$more = [];
	for (var ii = 0, jj = $binds.length; ii < jj; ii++) {
		var $bind = $binds[ii];
		var $field = $prototype.$properties[$bind];
		if ($field && !$field.$isExcluded) {
			var $item = $defined[$bind] ? _helpers.object.clone($defined[$bind], true) : {
				$bind: $bind
			};
			if ($field.$type == "application/x-array") {
				$array.push($item);
			} else {
				if (isCube) {
					if ($bind.indexOf("datend") >= 0) {
						continue;
					} else {
						if ($bind.indexOf("datstart") >= 0) {
							$item.$css = "s-field-cube-date";
							$more.push($item);
							$more.push({
								$bind: $bind.replace("datstart", "datend"),
								$css: "s-field-cube-date"
							});
							continue;
						}
					}
				}
				$more.push($item);
			}
		}
	}
	if ($more.length) {
		$article.$layout = {
			$items: $more.concat($array)
		};
	} else {
		$article.$layout.$items = $array;
	}
}

function _buildDefaultArticle($article, $prototype) {
	if (!$article.$layout.$items) {
		var $items = [];
		var binds = Object.keys($prototype.$properties);
		for (var ii = 0, jj = binds.length; ii < jj; ii++) {
			var $field = $prototype.$properties[binds[ii]];
			if ($field && !$field.$isExcluded) {
				$items.push({
					$bind: binds[ii]
				});
			}
		}
		$article.$layout = {
			$items: [{
				$category: "section",
				$layout: {
					$items: [{
						$category: "section",
						$layout: {
							$items: $items
						}
					}]
				}
			}]
		};
	}
}


function _ensureDefaultArticle(page, $article, $prototype) {
	if (!$article) {
		$article = {};
	}
	if (!$article.$layout) {
		$article.$layout = {};
	}
	switch (page.$facet) {
		case "$query":
		case "$bulk":
		case "$cube":
		case "$lookup":
		case "$select":
			_buildQueryArticle(page, $article, $prototype, page.$facet == "$cube");
			break;
		default:
			_buildDefaultArticle($article, $prototype);
			break;
	}
	return $article;
}


exports.ensureDefaultArticle = function(page, $article, $prototype) {
	if (page.ensureDefaultArticle) {
		return page.ensureDefaultArticle($article, $prototype);
	}
	return _ensureDefaultArticle(page, $article, $prototype);
};
