// syracuse-ui/lib/tests/pages/navigation/megamenu.js
"use strict";
require("./aim");
var _adllData = require("./navigationpage").data;
var _data = _adllData || {
	"$prototype": {
		"$type": "application/json",
		"$prototype": "{$baseUrl}/$prototypes('{$representation}.$details')",
		"$url": "{$baseUrl}/{$pluralType}('{$key}')?include=modules.code,modules.title,modules.submodules.code,modules.submodules.title",
		"$value": "{title}",
		"$title": "Navigation page",
		"$key": "{$uuid}",
		"$description": "Navigation page",
		"$pluralType": "navigationPages",
		"$representation": "navigationPage",
		"$baseUrl": "/sdata/syracuse/collaboration/syracuse",
		"$editMode": "user"
	},
	"$uuid": "f9daedd5-fd98-4461-b309-2b9bf796360a",
	"title": "Administration dashboard",
	"pageName": "home",
	"productName": " X3",
	"modules": [{
		"$uuid": "92269567-52f3-418f-a06f-19762dea5f4f",
		"$hostUrl": "http://gastonpc:8126",
		"$menuBaseUrl": "/sdata/syracuse/collaboration/syracuse",
		"code": "S_MOD_ADMIN",
		"title": "Administration",
		"submodules": [{
			"$uuid": "628e730a-d7f4-4347-99d4-8de49020d402",
			"code": "S_ADMIN",
			"title": "Administration",
			"items": [{
				"menuItem": {
					"$uuid": "daf5f986-b5eb-4b96-9d6d-23a1845e004a",
					"title": "Server logs",
					"description": "Server logs",
					"$links": {
						"$execute_$query": {
							"$type": "application/json;vnd.sage=syracuse",
							"$title": "Server logs",
							"$description": "Server logs",
							"$url": "{$menuBaseUrl}/serverLogs?representation=serverLog.$query",
							"$method": "GET",
							"$target": "self"
						},
						"$edit": {}
					},
					"entity": "serverLogs",
					"representation": "serverLog"
				}
			}, {
				"menuBlock": {
					"$uuid": "d54b8f76-6f2d-413e-86a0-f5c5f918aeff",
					"title": "Users",
					"items": [{
						"menuItem": {
							"$uuid": "234de8fb-57cf-4333-8956-d608d215734a",
							"title": "Users",
							"description": "Users management interface",
							"$links": {
								"$execute_$query": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "Users",
									"$description": "Users management interface",
									"$url": "{$menuBaseUrl}/users?representation=user.$query",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "users",
							"representation": "user"
						}
					}, {
						"menuItem": {
							"$uuid": "a3b6ea6c-c980-4e1a-b1e7-ff700523bdcc",
							"title": "Groups",
							"description": "Groups management",
							"$links": {
								"$execute_$query": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "Groups",
									"$description": "Groups management",
									"$url": "{$menuBaseUrl}/groups?representation=group.$query",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "groups",
							"representation": "group"
						}
					}, {
						"menuItem": {
							"$uuid": "8c21528d-cc97-42be-ade6-34f0ea1e09ad",
							"title": "Roles",
							"description": "User roles management interface",
							"$links": {
								"$execute_$query": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "Roles",
									"$description": "User roles management interface",
									"$url": "{$menuBaseUrl}/roles?representation=role.$query",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "roles",
							"representation": "role"
						}
					}, {
						"menuItem": {
							"$uuid": "61d7bb36-9698-4969-93bd-957b05ee8343",
							"title": "Security profiles",
							"description": "Security profiles",
							"$links": {
								"$execute_$query": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "Security profiles",
									"$description": "Security profiles",
									"$url": "{$menuBaseUrl}/securityProfiles?representation=securityProfile.$query",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "securityProfiles",
							"representation": "securityProfile"
						}
					}, {
						"menuItem": {
							"$uuid": "f48fc216-11e3-4c1f-b4c0-0415478abe2d",
							"title": "BO Profiles",
							"description": null,
							"$links": {
								"$execute_$query": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "BO Profiles",
									"$description": null,
									"$url": "{$menuBaseUrl}/boProfiles?representation=boProfile.$query",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "boProfiles",
							"representation": "boProfile"
						}
					}],
					"$links": {
						"$edit": {}
					}
				}
			}, {
				"menuBlock": {
					"$uuid": "e055f629-54f4-4cac-b5a8-ed938f805664",
					"title": "Settings",
					"items": [{
						"menuItem": {
							"$uuid": "27ef47f6-ee4b-418d-8a2f-b32597585e04",
							"title": "Global settings",
							"description": "Global settings management interface",
							"$links": {
								"$execute_$details": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "Global settings",
									"$description": "Global settings management interface",
									"$url": "{$menuBaseUrl}/settings(code eq 'settings')?representation=setting.$details",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "settings",
							"representation": "setting"
						}
					}, {
						"menuItem": {
							"$uuid": "13d182ef-dc57-4737-b563-4142632925ca",
							"title": "Locales",
							"description": "Locales management interface",
							"$links": {
								"$execute_$query": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "Locales",
									"$description": "Locales management interface",
									"$url": "{$menuBaseUrl}/localePreferences?representation=localePreference.$query",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "localePreferences",
							"representation": "localePreference"
						}
					}, {
						"menuItem": {
							"$uuid": "43b0a458-36bd-4395-b375-c483de7b4689",
							"title": "Proxies configuration",
							"description": "Proxies configuration interface",
							"$links": {
								"$execute_$query": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "Proxies configuration",
									"$description": "Proxies configuration interface",
									"$url": "{$menuBaseUrl}/proxyConfigurations?representation=proxyConfiguration.$query",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "proxyConfigurations",
							"representation": "proxyConfiguration"
						}
					}, {
						"menuBlock": {
							"$uuid": "2f7a1e08-b351-4b3c-9b0c-47556f087417",
							"title": "Authentication",
							"items": [{
								"menuItem": {
									"$uuid": "d6affc29-40d1-41bd-b209-f4f2160401ae",
									"title": "LDAP Servers",
									"description": "LDAP Servers management interface",
									"$links": {
										"$execute_$query": {
											"$type": "application/json;vnd.sage=syracuse",
											"$title": "LDAP Servers",
											"$description": "LDAP Servers management interface",
											"$url": "{$menuBaseUrl}/ldaps?representation=ldap.$query",
											"$method": "GET",
											"$target": "self"
										},
										"$edit": {
											"$isHidden": true
										}
									},
									"entity": "ldaps",
									"representation": "ldap"
								}
							}, {
								"menuItem": {
									"$uuid": "4d9e892a-82cf-4759-961c-4ae6da834577",
									"title": "OAuth2 Servers",
									"description": "OAuth2 Servers management interface",
									"$links": {
										"$execute_$query": {
											"$type": "application/json;vnd.sage=syracuse",
											"$title": "OAuth2 Servers",
											"$description": "OAuth2 Servers management interface",
											"$url": "{$menuBaseUrl}/oauth2s?representation=oauth2.$query",
											"$method": "GET",
											"$target": "self"
										},
										"$edit": {
											"$isHidden": true
										}
									},
									"entity": "oauth2s",
									"representation": "oauth2"
								}
							}],
							"$links": {
								"$edit": {}
							}
						}
					}, {
						"menuBlock": {
							"$uuid": "36e79c59-c49d-43be-9610-4ffd3abf53a3",
							"title": "Notifications",
							"items": [{
								"menuItem": {
									"$uuid": "9a0fc7d3-d47b-4ed7-90e2-99a64fe259f3",
									"title": "Notification servers",
									"description": "Notification servers",
									"$links": {
										"$execute_$query": {
											"$type": "application/json;vnd.sage=syracuse",
											"$title": "Notification servers",
											"$description": "Notification servers",
											"$url": "{$menuBaseUrl}/notificationServers?representation=notificationServer.$query",
											"$method": "GET",
											"$target": "self"
										},
										"$edit": {
											"$isHidden": true
										}
									},
									"entity": "notificationServers",
									"representation": "notificationServer"
								}
							}, {
								"menuItem": {
									"$uuid": "2fe5cf3c-664e-4585-b4c9-8293d17710d1",
									"title": "Notification events",
									"description": "Notification events",
									"$links": {
										"$execute_$query": {
											"$type": "application/json;vnd.sage=syracuse",
											"$title": "Notification events",
											"$description": "Notification events",
											"$url": "{$menuBaseUrl}/notificationEvents?representation=notificationEvent.$query",
											"$method": "GET",
											"$target": "self"
										},
										"$edit": {
											"$isHidden": true
										}
									},
									"entity": "notificationEvents",
									"representation": "notificationEvent"
								}
							}, {
								"menuItem": {
									"$uuid": "9eeffd73-15a7-4e17-b13d-33507d6c8fd0",
									"title": "Notification themes",
									"description": "Notification themes",
									"$links": {
										"$execute_$query": {
											"$type": "application/json;vnd.sage=syracuse",
											"$title": "Notification themes",
											"$description": "Notification themes",
											"$url": "{$menuBaseUrl}/notificationThemes?representation=notificationTheme.$query",
											"$method": "GET",
											"$target": "self"
										},
										"$edit": {
											"$isHidden": true
										}
									},
									"entity": "notificationThemes",
									"representation": "notificationTheme"
								}
							}, {
								"menuItem": {
									"$uuid": "a69d8c28-1ead-412e-86ab-877febac628b",
									"title": "Notification templates",
									"description": "Notification templates",
									"$links": {
										"$execute_$query": {
											"$type": "application/json;vnd.sage=syracuse",
											"$title": "Notification templates",
											"$description": "Notification templates",
											"$url": "{$menuBaseUrl}/mailTemplates?representation=mailTemplate.$query",
											"$method": "GET",
											"$target": "self"
										},
										"$edit": {
											"$isHidden": true
										}
									},
									"entity": "mailTemplates",
									"representation": "mailTemplate"
								}
							}],
							"$links": {
								"$edit": {}
							}
						}
					}],
					"$links": {
						"$edit": {}
					}
				}
			}, {
				"menuBlock": {
					"$uuid": "7a741bd4-b990-4de8-bab8-f08be54f462d",
					"title": "Endpoints",
					"$r_url": "{$r_url}/items('7a741bd4-b990-4de8-bab8-f08be54f462d')",
					"items": [{
						"menuItem": {
							"$uuid": "d13556f7-5d15-439e-9dbb-93b8ebbb7d39",
							"title": "Applications",
							"description": "Applications management interface",
							"$links": {
								"$execute_$query": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "Applications",
									"$description": "Applications management interface",
									"$url": "{$menuBaseUrl}/applications?representation=application.$query",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "applications",
							"representation": "application"
						}
					}, {
						"menuItem": {
							"$uuid": "8a761057-1f59-4f43-9434-1af6dd6d79ae",
							"title": "Endpoints",
							"description": "Service access endpoints management",
							"$links": {
								"$execute_$query": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "Endpoints",
									"$description": "Service access endpoints management",
									"$url": "{$menuBaseUrl}/endPoints?representation=endPoint.$query",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "endPoints",
							"representation": "endPoint"
						}
					}],
					"$links": {
						"$edit": {}
					}
				}
			}, {
				"menuBlock": {
					"$uuid": "6b93c6be-1880-48e0-8448-5257ffe6b31f",
					"title": "Licenses",
					"$r_url": "{$r_url}/items('6b93c6be-1880-48e0-8448-5257ffe6b31f')",
					"items": [{
						"menuItem": {
							"$uuid": "e8ea2290-14d7-42e4-85e7-d3953d9da745",
							"title": "About",
							"description": null,
							"$links": {
								"$execute_$details": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "About",
									"$description": null,
									"$url": "{$menuBaseUrl}/abouts('ALL')?representation=about.$details",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "abouts",
							"representation": "about"
						}
					}, {
						"menuItem": {
							"$uuid": "2be9155e-22fa-4f51-b215-1f3a0c72b7ec",
							"title": "License data",
							"description": "License data",
							"$links": {
								"$execute_$query": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "License data",
									"$description": "License data",
									"$url": "{$menuBaseUrl}/licenseDatas?representation=licenseData.$query",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "licenseDatas",
							"representation": "licenseData"
						}
					}, {
						"menuItem": {
							"$uuid": "59fe9e45-b131-494c-8afa-466474edd7af",
							"title": "Badges",
							"description": "Badges",
							"$links": {
								"$execute_$query": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "Badges",
									"$description": "Badges",
									"$url": "{$menuBaseUrl}/badges?representation=badge.$query",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "badges",
							"representation": "badge"
						}
					}, {
						"menuItem": {
							"$uuid": "a374c36c-4049-44a0-b63c-dc5f64ec38e8",
							"title": "License upload",
							"description": "License management",
							"$links": {
								"$execute_$edit": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "License upload",
									"$description": "License management",
									"$url": "{$menuBaseUrl}/licenses/$template/$workingCopies?representation=license.$edit&$method=POST",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "licenses",
							"representation": "license"
						}
					}],
					"$links": {
						"$edit": {}
					}
				}
			}, {
				"menuBlock": {
					"$uuid": "489ea83e-e1a3-44fe-b2da-bbe06598a61f",
					"title": "Certificates",
					"$r_url": "{$r_url}/items('489ea83e-e1a3-44fe-b2da-bbe06598a61f')",
					"items": [{
						"menuItem": {
							"$uuid": "8caf1e53-8b68-46fb-bbd9-f22981c9c8ad",
							"title": "Certificates",
							"description": "Certificates",
							"$links": {
								"$execute_$query": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "Certificates",
									"$description": "Certificates",
									"$url": "{$menuBaseUrl}/certificates?representation=certificate.$query",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "certificates",
							"representation": "certificate"
						}
					}, {
						"menuItem": {
							"$uuid": "85661061-b30b-4eb6-ba94-19a6f0cca938",
							"title": "Certificates of Certification Authorities",
							"description": "Certificates of Certification Authorities",
							"$links": {
								"$execute_$query": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "Certificates of Certification Authorities",
									"$description": "Certificates of Certification Authorities",
									"$url": "{$menuBaseUrl}/caCertificates?representation=caCertificate.$query",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "caCertificates",
							"representation": "caCertificate"
						}
					}],
					"$links": {
						"$edit": {}
					}
				}
			}, {
				"menuBlock": {
					"$uuid": "6cc0f2e0-2c54-4a85-8710-923e96fe0ed3",
					"title": "Servers",
					"$r_url": "{$r_url}/items('6cc0f2e0-2c54-4a85-8710-923e96fe0ed3')",
					"items": [{
						"menuItem": {
							"$uuid": "8158e8c6-226b-49d4-b064-089b6d680e6c",
							"title": "Hosts",
							"description": "Hosts",
							"$links": {
								"$execute_$query": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "Hosts",
									"$description": "Hosts",
									"$url": "{$menuBaseUrl}/hosts?representation=host.$query",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "hosts",
							"representation": "host"
						}
					}, {
						"menuItem": {
							"$uuid": "f1814d5d-8a49-4252-8db7-2f50e81b95e3",
							"title": "X3 solutions",
							"description": "X3 solutions",
							"$links": {
								"$execute_$query": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "X3 solutions",
									"$description": "X3 solutions",
									"$url": "{$menuBaseUrl}/x3solutions?representation=x3solution.$query",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "x3solutions",
							"representation": "x3solution"
						}
					}, {
						"menuItem": {
							"$uuid": "9f58addf-52c5-4200-9fa5-b90a3fdbd0a7",
							"title": "BO Servers",
							"description": null,
							"$links": {
								"$execute_$query": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "BO Servers",
									"$description": null,
									"$url": "{$menuBaseUrl}/boServers?representation=boServer.$query",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "boServers",
							"representation": "boServer"
						}
					}, {
						"menuBlock": {
							"$uuid": "87e93d6f-d2ef-4b69-b469-7c694eff882c",
							"title": "HRM Portal",
							"$r_url": "{$r_url}/items('87e93d6f-d2ef-4b69-b469-7c694eff882c')",
							"items": [{
								"menuItem": {
									"$uuid": "350baeeb-305a-4cb4-95c9-92ef0805032f",
									"title": "HRM Servers",
									"description": "HRM Servers",
									"$links": {
										"$execute_$query": {
											"$type": "application/json;vnd.sage=syracuse",
											"$title": "HRM Servers",
											"$description": "HRM Servers",
											"$url": "{$menuBaseUrl}/hrmServers?representation=hrmServer.$query",
											"$method": "GET",
											"$target": "self"
										},
										"$edit": {
											"$isHidden": true
										}
									},
									"entity": "hrmServers",
									"representation": "hrmServer"
								}
							}, {
								"menuItem": {
									"$uuid": "ac102882-d256-4c1e-a15e-713c550afa83",
									"title": "HRM Sites",
									"description": "HRM Sites",
									"$links": {
										"$execute_$query": {
											"$type": "application/json;vnd.sage=syracuse",
											"$title": "HRM Sites",
											"$description": "HRM Sites",
											"$url": "{$menuBaseUrl}/hrmSites?representation=hrmSite.$query",
											"$method": "GET",
											"$target": "self"
										},
										"$edit": {
											"$isHidden": true
										}
									},
									"entity": "hrmSites",
									"representation": "hrmSite"
								}
							}],
							"$links": {
								"$edit": {}
							}
						}
					}],
					"$links": {
						"$edit": {}
					}
				}
			}, {
				"menuBlock": {
					"$uuid": "e6d6ba7a-c3ef-43ef-aeef-fd914b5212d6",
					"title": "Web Services",
					"$r_url": "{$r_url}/items('e6d6ba7a-c3ef-43ef-aeef-fd914b5212d6')",
					"items": [{
						"menuItem": {
							"$uuid": "fa07db47-2005-4fbe-bbd5-9e1d2ea205de",
							"title": "Classic SOAP Web services",
							"description": "Classic SOAP Web services",
							"$links": {
								"$execute_$query": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "Classic SOAP Web services",
									"$description": "Classic SOAP Web services",
									"$url": "{$menuBaseUrl}/soapGenerics?representation=soapGeneric.$query",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "soapGenerics",
							"representation": "soapGeneric"
						}
					}, {
						"menuItem": {
							"$uuid": "51faf721-36b9-4ced-ac83-0a65de6bfffe",
							"title": "Classic SOAP pools configuration",
							"description": "Classic SOAP pools configuration",
							"$links": {
								"$execute_$query": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "Classic SOAP pools configuration",
									"$description": "Classic SOAP pools configuration",
									"$url": "{$menuBaseUrl}/soapClassicPools?representation=soapClassicPool.$query",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "soapClassicPools",
							"representation": "soapClassicPool"
						}
					}, {
						"menuItem": {
							"$uuid": "2a66d835-5713-4c1f-81fa-b801537e28bd",
							"title": "Rest web services",
							"description": null,
							"$links": {
								"$execute_$query": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "Rest web services",
									"$description": null,
									"$url": "{$menuBaseUrl}/restWebServices?representation=restWebService.$query",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "restWebServices",
							"representation": "restWebService"
						}
					}],
					"$links": {
						"$edit": {}
					}
				}
			}],
			"$links": {
				"$edit": {
					"$isHidden": true
				}
			}
		}, {
			"$uuid": "b23b1683-8d56-4323-b428-74b30026ca57",
			"$r_url": "/sdata/syracuse/collaboration/syracuse/menuBlocks('b23b1683-8d56-4323-b428-74b30026ca57')",
			"code": "S_COLL",
			"title": "Collaboration",
			"items": [{
				"menuItem": {
					"$uuid": "b3ca3f47-63c4-4c75-8ad7-a50627b907d4",
					"title": "Teams",
					"description": "Teams management",
					"$links": {
						"$execute_$query": {
							"$type": "application/json;vnd.sage=syracuse",
							"$title": "Teams",
							"$description": "Teams management",
							"$url": "{$menuBaseUrl}/teams?representation=team.$query",
							"$method": "GET",
							"$target": "self"
						},
						"$edit": {
							"$isHidden": true
						}
					},
					"entity": "teams",
					"representation": "team",
					"$capabilities": ""
				}
			}, {
				"menuItem": {
					"$uuid": "5d345818-1ecc-4c47-8a09-3b0e36d7f1e6",
					"title": "Volumes",
					"description": "Storage volumes",
					"$links": {
						"$execute_$query": {
							"$type": "application/json;vnd.sage=syracuse",
							"$title": "Volumes",
							"$description": "Storage volumes",
							"$url": "{$menuBaseUrl}/storageVolumes?representation=storageVolume.$query",
							"$method": "GET",
							"$target": "self"
						},
						"$edit": {
							"$isHidden": true
						}
					},
					"entity": "storageVolumes",
					"representation": "storageVolume",
					"$capabilities": ""
				}
			}, {
				"menuItem": {
					"$uuid": "b9771625-b70d-4a7b-825c-cbbfb9a3b272",
					"title": "Documents",
					"description": "Document management",
					"$links": {
						"$execute_$query": {
							"$type": "application/json;vnd.sage=syracuse",
							"$title": "Documents",
							"$description": "Document management",
							"$url": "{$menuBaseUrl}/documents?representation=document.$query",
							"$method": "GET",
							"$target": "self"
						},
						"$edit": {
							"$isHidden": true
						}
					},
					"entity": "documents",
					"representation": "document",
					"$capabilities": ""
				}
			}, {
				"menuItem": {
					"$uuid": "7f40d8c6-1749-4a2c-be7a-b458e1de0d28",
					"title": "Document tags",
					"description": "Document tags",
					"$links": {
						"$execute_$query": {
							"$type": "application/json;vnd.sage=syracuse",
							"$title": "Document tags",
							"$description": "Document tags",
							"$url": "{$menuBaseUrl}/documentTags?representation=documentTag.$query",
							"$method": "GET",
							"$target": "self"
						},
						"$edit": {
							"$isHidden": true
						}
					},
					"entity": "documentTags",
					"representation": "documentTag",
					"$capabilities": ""
				}
			}, {
				"menuItem": {
					"$uuid": "c97e4a5c-1a4b-4de8-a062-04d8bbe48a0e",
					"title": "Tags categories",
					"description": "Tags categories",
					"$links": {
						"$execute_$query": {
							"$type": "application/json;vnd.sage=syracuse",
							"$title": "Tags categories",
							"$description": "Tags categories",
							"$url": "{$menuBaseUrl}/documentTagCategories?representation=documentTagCategory.$query",
							"$method": "GET",
							"$target": "self"
						},
						"$edit": {
							"$isHidden": true
						}
					},
					"entity": "documentTagCategories",
					"representation": "documentTagCategory",
					"$capabilities": ""
				}
			}, {
				"menuItem": {
					"$uuid": "ab9b311f-c501-4acb-ab2d-a97d7edf9aa3",
					"title": "Word templates",
					"description": "Word templates",
					"$links": {
						"$execute_$query": {
							"$type": "application/json;vnd.sage=syracuse",
							"$title": "Word templates",
							"$description": "Word templates",
							"$url": "{$menuBaseUrl}/msoWordTemplateDocuments?representation=msoWordTemplateDocument.$query",
							"$method": "GET",
							"$target": "self"
						},
						"$edit": {
							"$isHidden": true
						}
					},
					"entity": "msoWordTemplateDocuments",
					"representation": "msoWordTemplateDocument",
					"$capabilities": ""
				}
			}],
			"$links": {
				"$edit": {
					"$isHidden": true
				}
			}
		}, {
			"$uuid": "ad0e8631-f788-49b8-b066-012961cefbcd",
			"$r_url": "/sdata/syracuse/collaboration/syracuse/menuBlocks('ad0e8631-f788-49b8-b066-012961cefbcd')",
			"code": "S_PERS",
			"title": "Authoring",
			"items": [{
				"menuItem": {
					"$uuid": "2b218545-7f42-459d-add1-43c28928cde1",
					"title": "S_THEMES",
					"description": "",
					"$links": {
						"$execute_$query": {
							"$type": "application/json;vnd.sage=syracuse",
							"$title": "S_THEMES",
							"$description": "",
							"$url": "{$menuBaseUrl}/themes?representation=theme.$query",
							"$method": "GET",
							"$target": "self"
						},
						"$edit": {}
					},
					"entity": "themes",
					"representation": "theme"
				}
			}, {
				"menuItem": {
					"$uuid": "3d5c2dd0-a358-4e26-8d07-1aa19f1a2a71",
					"title": "Navigation pages",
					"description": null,
					"$links": {
						"$execute_$query": {
							"$type": "application/json;vnd.sage=syracuse",
							"$title": "Navigation pages",
							"$description": null,
							"$url": "{$menuBaseUrl}/navigationPages?representation=navigationPage.$query",
							"$method": "GET",
							"$target": "self"
						},
						"$edit": {
							"$isHidden": true
						}
					},
					"entity": "navigationPages",
					"representation": "navigationPage",
					"$capabilities": ""
				}
			}, {
				"menuItem": {
					"$uuid": "1846c78b-a466-4296-b471-cf0eb78b4f4e",
					"title": "Menu Modules",
					"description": null,
					"$links": {
						"$execute_$query": {
							"$type": "application/json;vnd.sage=syracuse",
							"$title": "Menu Modules",
							"$description": null,
							"$url": "{$menuBaseUrl}/menuModules?representation=menuModule.$query",
							"$method": "GET",
							"$target": "self"
						},
						"$edit": {
							"$isHidden": true
						}
					},
					"entity": "menuModules",
					"representation": "menuModule",
					"$capabilities": ""
				}
			}, {
				"menuItem": {
					"$uuid": "159d141c-9aa6-4ee1-b458-3d2d3ba6d379",
					"title": "Menu Submodules",
					"description": null,
					"$links": {
						"$execute_$query": {
							"$type": "application/json;vnd.sage=syracuse",
							"$title": "Menu Submodules",
							"$description": null,
							"$url": "{$menuBaseUrl}/menuBlocks?representation=menuBlock.$query",
							"$method": "GET",
							"$target": "self"
						},
						"$edit": {
							"$isHidden": true
						}
					},
					"entity": "menuBlocks",
					"representation": "menuBlock",
					"$capabilities": ""
				}
			}, {
				"menuItem": {
					"$uuid": "fa0cd13e-6fba-42e2-bab2-48fc63a19f66",
					"title": "Home pages",
					"description": null,
					"$links": {
						"$execute_$query": {
							"$type": "application/json;vnd.sage=syracuse",
							"$title": "Home pages",
							"$description": null,
							"$url": "{$menuBaseUrl}/landingPages?representation=landingPage.$query",
							"$method": "GET",
							"$target": "self"
						},
						"$edit": {
							"$isHidden": true
						}
					},
					"entity": "landingPages",
					"representation": "landingPage",
					"$capabilities": ""
				}
			}, {
				"menuItem": {
					"$uuid": "ebde23b7-7942-4b84-8a23-594b5c010f2d",
					"title": "Menu Items",
					"description": "Menu items management interface",
					"$links": {
						"$execute_$query": {
							"$type": "application/json;vnd.sage=syracuse",
							"$title": "Menu Items",
							"$description": "Menu items management interface",
							"$url": "{$menuBaseUrl}/menuItems?representation=menuItem.$query",
							"$method": "GET",
							"$target": "self"
						},
						"$edit": {
							"$isHidden": true
						}
					},
					"entity": "menuItems",
					"representation": "menuItem",
					"$capabilities": ""
				}
			}, {
				"menuItem": {
					"$uuid": "ca835319-466c-4f49-a0f1-00092983d7e9",
					"title": "Menu Categories",
					"description": "Menu categories management interface",
					"$links": {
						"$execute_$query": {
							"$type": "application/json;vnd.sage=syracuse",
							"$title": "Menu Categories",
							"$description": "Menu categories management interface",
							"$url": "{$menuBaseUrl}/menuCategories?representation=menuCategory.$query",
							"$method": "GET",
							"$target": "self"
						},
						"$edit": {
							"$isHidden": true
						}
					},
					"entity": "menuCategories",
					"representation": "menuCategory",
					"$capabilities": ""
				}
			}, {
				"menuItem": {
					"$uuid": "87b4d20c-afcb-4278-b2f8-8ed975de2c1c",
					"title": "Customized pages",
					"description": "Pages management interface",
					"$links": {
						"$execute_$query": {
							"$type": "application/json;vnd.sage=syracuse",
							"$title": "Customized pages",
							"$description": "Pages management interface",
							"$url": "{$menuBaseUrl}/pageDefs?representation=pageDef.$query",
							"$method": "GET",
							"$target": "self"
						},
						"$edit": {
							"$isHidden": true
						}
					},
					"entity": "pageDefs",
					"representation": "pageDef",
					"$capabilities": ""
				}
			}],
			"$links": {
				"$edit": {
					"$isHidden": true
				}
			}
		}, {
			"$uuid": "1f5ac326-506d-4a25-82b7-21296e3d76b4",
			"$r_url": "/sdata/syracuse/collaboration/syracuse/menuBlocks('1f5ac326-506d-4a25-82b7-21296e3d76b4')",
			"code": "S_TOOLS",
			"title": "Utilities",
			"items": [{
				"menuBlock": {
					"$uuid": "8759d172-49c7-4d1d-ae36-7ddd5635356a",
					"title": "Imports",
					"$r_url": "{$r_url}/items('8759d172-49c7-4d1d-ae36-7ddd5635356a')",
					"items": [{
						"menuItem": {
							"$uuid": "28f02ec9-c04b-4cf7-a16c-59fe21b7dac4",
							"title": "Import Tool",
							"description": "Import tool",
							"$links": {
								"$execute_$edit": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "Import Tool",
									"$description": "Import tool",
									"$url": "{$menuBaseUrl}/importTools/$template/$workingCopies?representation=importTool.$edit&$method=POST",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "importTools",
							"representation": "importTool"
						}
					}, {
						"menuItem": {
							"$uuid": "a5e221df-a3fc-4af1-b276-81bd6a9f5a05",
							"title": "X3 Users import",
							"description": "X3 Users import",
							"$links": {
								"$execute_$query": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "X3 Users import",
									"$description": "X3 Users import",
									"$url": "{$menuBaseUrl}/x3UserImports?representation=x3UserImport.$query",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "x3UserImports",
							"representation": "x3UserImport"
						}
					}, {
						"menuItem": {
							"$uuid": "3e836efb-2045-4df7-9a2a-4a80533d6938",
							"title": "Menu profile import",
							"description": null,
							"$links": {
								"$execute_$edit": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "Menu profile import",
									"$description": null,
									"$url": "{$menuBaseUrl}/profileMenuImports/$template/$workingCopies?representation=profileMenuImport.$edit&$method=POST",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "profileMenuImports",
							"representation": "profileMenuImport"
						}
					}],
					"$links": {
						"$edit": {}
					}
				}
			}, {
				"menuBlock": {
					"$uuid": "8919ca0c-f8ae-4264-8c38-14e12807cc42",
					"title": "Exports",
					"$r_url": "{$r_url}/items('8919ca0c-f8ae-4264-8c38-14e12807cc42')",
					"items": [{
						"menuItem": {
							"$uuid": "3f8220f5-cd75-4359-b86b-ae92ceb5549b",
							"title": "Export profiles",
							"description": "Export profiles",
							"$links": {
								"$execute_$query": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "Export profiles",
									"$description": "Export profiles",
									"$url": "{$menuBaseUrl}/exportProfiles?representation=exportProfile.$query",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "exportProfiles",
							"representation": "exportProfile"
						}
					}, {
						"menuItem": {
							"$uuid": "4315b554-2ed2-432a-870b-fa490f94eab7",
							"title": "Personalizations management",
							"description": null,
							"$links": {
								"$execute_$query": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "Personalizations management",
									"$description": null,
									"$url": "{$menuBaseUrl}/personalizationManagements?representation=personalizationManagement.$query",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "personalizationManagements",
							"representation": "personalizationManagement"
						}
					}, {
						"menuItem": {
							"$uuid": "a3e92f61-ef57-4389-92cc-c6470062fbd0",
							"title": "Resources packs",
							"description": null,
							"$links": {
								"$execute_$query": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "Resources packs",
									"$description": null,
									"$url": "{$menuBaseUrl}/resourcePacks?representation=resourcePack.$query",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "resourcePacks",
							"representation": "resourcePack"
						}
					}],
					"$links": {
						"$edit": {}
					}
				}
			}, {
				"menuBlock": {
					"$uuid": "76ce13e2-600e-49e5-920c-736af69411c7",
					"title": "Installation",
					"$r_url": "{$r_url}/items('76ce13e2-600e-49e5-920c-736af69411c7')",
					"items": [{
						"menuItem": {
							"$uuid": "a1888ff8-de23-4935-9a11-2d6e5a1cca80",
							"title": "Updates",
							"description": "\"One-click\" updates",
							"$links": {
								"$execute_$query": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "Updates",
									"$description": "\"One-click\" updates",
									"$url": "{$menuBaseUrl}/updates?representation=update.$query",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "updates",
							"representation": "update"
						}
					}, {
						"menuItem": {
							"$uuid": "6228bfe1-3a70-4992-aaf9-1ef80255af36",
							"title": "Install addins for Office",
							"description": null,
							"$links": {
								"$execute_$edit": {
									"$type": "html",
									"$title": "Install addins for Office",
									"$description": null,
									"$url": "/msoffice/lib/general/addIn/SyracuseOfficeAddinsSetup.EXE",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"externalUrl": "/msoffice/lib/general/addIn/SyracuseOfficeAddinsSetup.EXE"
						}
					}, {
						"menuItem": {
							"$uuid": "434c6896-ad41-4f6a-a80b-1f44405b2247",
							"title": "Install addin for Outlook",
							"description": null,
							"$links": {
								"$execute_$edit": {
									"$type": "html",
									"$title": "Install addin for Outlook",
									"$description": null,
									"$url": "/msoffice/lib/general/addin/SyracuseOutlookAddinSetup.EXE",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"externalUrl": "/msoffice/lib/general/addin/SyracuseOutlookAddinSetup.EXE"
						}
					}, {
						"menuItem": {
							"$uuid": "a8fa2f67-4c5b-4250-bf00-b2b0028d424e",
							"title": "Ilog Connector",
							"description": "Connector for Scales and Gantt ILOG modules",
							"$links": {
								"$execute_$edit": {
									"$type": "html",
									"$title": "Ilog Connector",
									"$description": "Connector for Scales and Gantt ILOG modules",
									"$url": "/plugin/download/safex3-ilogconnector-installer-x86_64.exe",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"externalUrl": "/plugin/download/safex3-ilogconnector-installer-x86_64.exe"
						}
					}, {
						"menuItem": {
							"$uuid": "65cc27e9-f6e8-4519-97da-c5141a4b195b",
							"title": "Crystal Report Connector",
							"description": "Connector for Crystal Report developer mode",
							"$links": {
								"$execute_$edit": {
									"$type": "html",
									"$title": "Crystal Report Connector",
									"$description": "Connector for Crystal Report developer mode",
									"$url": "/plugin/download/safex3-crconnector-installer-x86_64.exe",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"externalUrl": "/plugin/download/safex3-crconnector-installer-x86_64.exe"
						}
					}],
					"$links": {
						"$edit": {}
					}
				}
			}],
			"$links": {
				"$edit": {
					"$isHidden": true
				}
			}
		}, {
			"$uuid": "b8df68cd-ce77-4923-a496-b640931ed6e3",
			"$r_url": "/sdata/syracuse/collaboration/syracuse/menuBlocks('b8df68cd-ce77-4923-a496-b640931ed6e3')",
			"code": "S_MOBILITY",
			"title": "Mobile applications",
			"items": [{
				"menuItem": {
					"$uuid": "b65d22bd-9064-447e-9986-f3bd2d524189",
					"title": "Mobile applications",
					"description": "Mobile applications",
					"$links": {
						"$execute_$query": {
							"$type": "application/json;vnd.sage=syracuse",
							"$title": "Mobile applications",
							"$description": "Mobile applications",
							"$url": "{$menuBaseUrl}/mobileApplications?representation=mobileApplication.$query",
							"$method": "GET",
							"$target": "self"
						},
						"$edit": {
							"$isHidden": true
						}
					},
					"entity": "mobileApplications",
					"representation": "mobileApplication",
					"$capabilities": ""
				}
			}, {
				"menuItem": {
					"$uuid": "55025d7c-4cd3-458e-a8e1-49a42568cb83",
					"title": "Mobile dashboards",
					"description": "Mobile dashboards",
					"$links": {
						"$execute_$query": {
							"$type": "application/json;vnd.sage=syracuse",
							"$title": "Mobile dashboards",
							"$description": "Mobile dashboards",
							"$url": "{$menuBaseUrl}/mobileDashboards?representation=mobileDashboard.$query",
							"$method": "GET",
							"$target": "self"
						},
						"$edit": {
							"$isHidden": true
						}
					},
					"entity": "mobileDashboards",
					"representation": "mobileDashboard",
					"$capabilities": ""
				}
			}, {
				"menuItem": {
					"$uuid": "592240f7-a242-4237-9c51-aa3ec17ac307",
					"title": "Mobile gadgets",
					"description": "Mobile gadgets",
					"$links": {
						"$execute_$query": {
							"$type": "application/json;vnd.sage=syracuse",
							"$title": "Mobile gadgets",
							"$description": "Mobile gadgets",
							"$url": "{$menuBaseUrl}/mobileGadgets?representation=mobileGadget.$query",
							"$method": "GET",
							"$target": "self"
						},
						"$edit": {
							"$isHidden": true
						}
					},
					"entity": "mobileGadgets",
					"representation": "mobileGadget",
					"$capabilities": ""
				}
			}, {
				"menuItem": {
					"$uuid": "d609f490-9fb2-472e-b583-154e9a88699a",
					"title": "Migrate mobile dashboard",
					"description": "Migrate mobile dashboard",
					"$links": {
						"$execute_$edit": {
							"$type": "application/json;vnd.sage=syracuse",
							"$title": "Migrate mobile dashboard",
							"$description": "Migrate mobile dashboard",
							"$url": "{$menuBaseUrl}/mobileMigrateDashboards/$template/$workingCopies?representation=mobileMigrateDashboard.$edit&$method=POST",
							"$method": "GET",
							"$target": "self"
						},
						"$edit": {
							"$isHidden": true
						}
					},
					"entity": "mobileMigrateDashboards",
					"representation": "mobileMigrateDashboard",
					"$capabilities": ""
				}
			}],
			"$links": {
				"$edit": {
					"$isHidden": true
				}
			}
		}, {
			"$uuid": "3bb36d15-7d2c-476b-9b04-ef2fb2c787de",
			"$r_url": "/sdata/syracuse/collaboration/syracuse/menuBlocks('3bb36d15-7d2c-476b-9b04-ef2fb2c787de')",
			"code": "S_USAGE",
			"title": "Usage",
			"items": [{
				"menuItem": {
					"$uuid": "fc6aaa4d-7934-4c16-85cd-13d754f4318b",
					"title": "Search Index Management",
					"description": "Search Index Management",
					"$links": {
						"$execute_$edit": {
							"$type": "application/json;vnd.sage=syracuse",
							"$title": "Search Index Management",
							"$description": "Search Index Management",
							"$url": "{$menuBaseUrl}/searchAdmins/$template/$workingCopies?representation=searchAdmin.$edit&$method=POST",
							"$method": "GET",
							"$target": "self"
						},
						"$edit": {
							"$isHidden": true
						}
					},
					"entity": "searchAdmins",
					"representation": "searchAdmin",
					"$capabilities": ""
				}
			}, {
				"menuBlock": {
					"$uuid": "c8f828c8-8a97-438a-bb4d-e61de877e732",
					"title": "Automate",
					"$r_url": "{$r_url}/items('c8f828c8-8a97-438a-bb4d-e61de877e732')",
					"items": [{
						"menuItem": {
							"$uuid": "692c782c-b203-46fc-bfe4-1069ee8c1c04",
							"title": "Scheduler",
							"description": "Scheduler",
							"$links": {
								"$execute_$query": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "Scheduler",
									"$description": "Scheduler",
									"$url": "{$menuBaseUrl}/automates?representation=automate.$query",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "automates",
							"representation": "automate"
						}
					}, {
						"menuItem": {
							"$uuid": "3b80c77f-20b6-47c8-9ca2-946a4dd5993c",
							"title": "Server logs",
							"description": "Server logs",
							"$links": {
								"$execute_$query": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "Server logs",
									"$description": "Server logs",
									"$url": "{$menuBaseUrl}/serverLogs?representation=serverLog.$query",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "serverLogs",
							"representation": "serverLog"
						}
					}, {
						"menuItem": {
							"$uuid": "82036412-8c87-4714-8ada-73210716e986",
							"title": "Import sessions",
							"description": "Import sessions",
							"$links": {
								"$execute_$query": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "Import sessions",
									"$description": "Import sessions",
									"$url": "{$menuBaseUrl}/importSessions?representation=importSession.$query",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "importSessions",
							"representation": "importSession"
						}
					}],
					"$links": {
						"$edit": {}
					}
				}
			}, {
				"menuBlock": {
					"$uuid": "3cfb22be-e3ea-40c3-9e23-342b2043a470",
					"title": "Logs",
					"$r_url": "{$r_url}/items('3cfb22be-e3ea-40c3-9e23-342b2043a470')",
					"items": [{
						"menuItem": {
							"$uuid": "37899552-67b5-4f91-89ae-674f3924391d",
							"title": "Host traces",
							"description": "Syracuse server traces",
							"$links": {
								"$execute_$query": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "Host traces",
									"$description": "Syracuse server traces",
									"$url": "{$menuBaseUrl}/traceRecords?representation=traceRecord.$query",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "traceRecords",
							"representation": "traceRecord"
						}
					}, {
						"menuItem": {
							"$uuid": "4c59424f-36cb-492a-a361-7b5648d0e67d",
							"title": "Engine trace",
							"description": null,
							"$links": {
								"$execute_$edit": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "Engine trace",
									"$description": null,
									"$url": "{$menuBaseUrl}/x3SessionConfigs/$template/$workingCopies?representation=x3SessionConfig.$edit&$method=POST",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "x3SessionConfigs",
							"representation": "x3SessionConfig"
						}
					}],
					"$links": {
						"$edit": {}
					}
				}
			}, {
				"menuBlock": {
					"$uuid": "dc73f211-2748-48ac-842e-e605ac0316fd",
					"title": "Sessions management",
					"$r_url": "{$r_url}/items('dc73f211-2748-48ac-842e-e605ac0316fd')",
					"items": [{
						"menuItem": {
							"$uuid": "3efc2b51-26ee-414f-a17d-059c828d903c",
							"title": "License follow up",
							"description": null,
							"$links": {
								"$execute_$details": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "License follow up",
									"$description": null,
									"$url": "{$menuBaseUrl}/licenseViews('all')?representation=licenseView.$details",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "licenseViews",
							"representation": "licenseView"
						}
					}, {
						"menuItem": {
							"$uuid": "f7ff1b2f-c4d3-4fb2-b673-dac9df49707e",
							"title": "Web client sessions",
							"description": "Sessions Management",
							"$links": {
								"$execute_$query": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "Web client sessions",
									"$description": "Sessions Management",
									"$url": "{$menuBaseUrl}/sessionInfos?representation=sessionInfo.$query",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "sessionInfos",
							"representation": "sessionInfo"
						}
					}, {
						"menuItem": {
							"$uuid": "ef2fffb5-3785-471e-bde0-95872acecfc9",
							"title": "Classic client sessions",
							"description": null,
							"$links": {
								"$execute_$query": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "Classic client sessions",
									"$description": null,
									"$url": "{$menuBaseUrl}/cvgSessions?representation=cvgSession.$query",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "cvgSessions",
							"representation": "cvgSession"
						}
					}, {
						"menuItem": {
							"$uuid": "e1af5660-73c1-445b-ae12-41e35098d29c",
							"title": "Clients reused list",
							"description": null,
							"$links": {
								"$execute_$query": {
									"$type": "application/json;vnd.sage=syracuse",
									"$title": "Clients reused list",
									"$description": null,
									"$url": "{$menuBaseUrl}/cvgReuseClients?representation=cvgReuseClient.$query",
									"$method": "GET",
									"$target": "self"
								},
								"$edit": {
									"$isHidden": true
								}
							},
							"entity": "cvgReuseClients",
							"representation": "cvgReuseClient"
						}
					}],
					"$links": {
						"$edit": {}
					}
				}
			}],
			"$links": {
				"$edit": {
					"$isHidden": true
				}
			}
		}],
		"$transMenuBaseUrl": "/trans/syracuse/collaboration/syracuse",
		"$links": {
			"$edit": {
				"$isHidden": true
			}
		}
	}],
	"$key": "f9daedd5-fd98-4461-b309-2b9bf796360a"
};

function _build() {
	var nav = document.createElement("nav");
	nav.className = "nav-menu";
	var ul = document.createElement("ul");
	ul.className = "clearfix";
	for (var ii = 0, jj = _data.modules.length; ii < jj; ii++) {
		ul.appendChild(_addModule(_data.modules[ii]));
	}
	nav.appendChild(ul);




	var wrapper = document.createElement("div");
	wrapper.className = "wrapper";
	wrapper.appendChild(nav);

	var headerMenu = document.createElement("div");
	headerMenu.className = "header";
	headerMenu.appendChild(wrapper);

	var header = document.createElement("header");
	header.className = "header";
	header.appendChild(headerMenu);
	return header;
};

function _addModule(module) {
	var li = document.createElement("li");
	li.className = "top-tier";
	var title = document.createElement("span");
	title.textContent = module.title;
	li.appendChild(title);

	var ul = document.createElement("ul");
	ul.className = "std-menu";
	for (var ii = 0, jj = module.submodules.length; ii < jj; ii++) {
		ul.appendChild(_addSubModule(module.submodules[ii]));
	}
	li.appendChild(ul);
	return li;
}



function _addSubModule(submodule) {
	var li = document.createElement("li");
	li.className = "menu-group";
	var title = document.createElement("span");
	title.textContent = submodule.title;
	li.appendChild(title);

	var slot = document.createElement("div");
	slot.className = "sub-menu-wrap";
	slot.style.display = "block";

	_addItems(slot, submodule.items);

	li.appendChild(slot);
	return li;
}

function _addItems(parentNode, items) {
	for (var ii = 0, jj = items.length; ii < jj; ii++) {
		var item = items[ii];
		if (item.menuItem) {
			var link = document.createElement("a");
			link.textContent = item.menuItem.title;
			var li = document.createElement("li");
			li.appendChild(link);
			parentNode.appendChild(li);
		} else {
			if (item.menuBlock) {
				var slot = document.createElement("div");
				slot.className = "menu-section";
				var li = document.createElement("li");
				li.className = "sub-heading";
				li.textContent = item.menuBlock.title;
				var ul = document.createElement("ul");
				ul.appendChild(li);
				_addItems(ul, item.menuBlock.items);
				slot.appendChild(ul);
				parentNode.appendChild(slot);
			}
		}
	}
}

function _useAim() {

	var $menu = $(".nav-menu .std-menu");
	$menu.menuAim({
		activate: activateSubmenu,
		deactivate: deactivateSubmenu,
		exitMenu: exitSubmenu
	});

	function activateSubmenu(row) {
		var $row = $(row),
			$submenu = $row.find(".sub-menu-wrap"); //,
		// Show the submenu
		$submenu.css({
			display: "block"
		});

		$row.find("span:first").addClass("active");
	}

	function deactivateSubmenu(row) {
		var $row = $(row),
			$submenu = $row.find(".sub-menu-wrap");

		// Hide the submenu and remove the row's highlighted look
		$submenu.css("display", "none");
		$row.find("span:first").removeClass("active");
	}

	function exitSubmenu(row) {
		var $row = $(row);
		$row.find(".sub-menu-wrap").hide().eq(0).show();
	}

	$(".nav-menu .std-menu li").click(function(e) {
		e.stopPropagation();
	});

	$menu.find(".menu-section li:not('.sub-heading')").click(function() {
		$(".std-menu").addClass("deactive").find("> li:not(:first-child) .sub-menu-wrap").css("display", "none");
		$(".nav-menu span.active").removeClass("active");
	});

	$(".nav-menu .top-tier").hover(function() {
		$(this).find(".active").removeClass("active");
		$(this).find("li:first span:first").addClass("active");
		$(this).find(".deactive").removeClass("deactive");
	}, function() {
		$(this).find("li:first span:first").removeClass("active");
	});

	$(".nav-secondary > ul > li:eq(1) a ").mouseenter().addClass("strong");

}

exports.load = function(page, finish) {
	syra_site.loadStyleSheet("/syracuse-ui/lib/tests/pages/navigation/megamenu.css", true);
	page.layoutContent.domItem.appendChild(_build());


	_useAim();

	finish(page);
};
