// syracuse-ui/lib/fusion/core/client/sap/srvactions.js
"use strict";

var _srvAction = {
	"SRV_FIN_APPLI": 51,
	"SRV_CHAMP_SUIVANT": 1,
	"SRV_BOITE_SELECTION": 2,
	"SRV_SORTIE_CHAMP_SPE": 3, // Obsolete?
	"SRV_VALIDE_ECRAN": 4,
	"SRV_CHAMP_PRECEDENT": 31,
	"SRV_ABANDON": 7,
	"SRV_ABANDONLINE": 6,
	"SRV_VALIDELINE": 23,
	"SRV_HELP": 12, // Obsolete?
	"SRV_DEBUGGER": 16,
	"SRV_CALCULATRICE": 17,
	"SRV_GETFOCUS": 20,
	"SRV_GETFOCUSCELL": 26, // Grid 
	"SRV_INSERTLINEGRID": 73,
	"SRV_DELETELINEGRID": 65,
	"SRV_DELETELINESGRID": 83,
	"SRV_COMMANDGRID": 21,
	"SRV_CUTLINEGRID": 68,
	"SRV_COPYLINEGRID": 66,
	"SRV_PASTELINEGRID": 67,
	"SRV_BRWCHANGELINE": 27, // Query Lists 
	"SRV_BRWSELECTLINE": 28,
	"SRV_BRWDESELECTLINE": 43,
	"SRV_BRWRAPIDESEL": 45,
	"SRV_BRWBEGINEDIT": 32, // Obsolete?
	"SRV_BRWENDEDIT": 33, // Obsolete?
	"SRV_CHAMP_ONGLET": 29,
	"SRV_ABANDON_SAISIE": 6, // Obsolete?
	"SRV_TREECHANGELINE": 36, // Tree Lists
	"SRV_TREESELECTLINE": 37,
	"SRV_TREEDESELECTLINE": 41,
	"SRV_TREEEXPANDLINE": 38,
	"SRV_TREEBEGINEDIT": 39, // Obsolete?
	"SRV_TREEENDEDIT": 40, // Obsolete?
	"SRV_PICKING": 42,
	"SRV_FERME_PICKING": 44,
	"SRV_CTL_FORMULE": 46,
	"SRV_ELEMENT_FORMULE": 47,
	"SRV_LIST_PAGE_UP": 48,
	"SRV_LIST_PAGE_DOWN": 49,
	"SRV_OPEN_CTXMENU": 50,
	"SRV_TIMEOUTSESS": 0,
	"SRV_EXEFNC": 52,
	"SRV_ASKUI": 56,
	"SRV_RESI": 57,
	"SRV_KIB": 178,
	"SRV_FICHIER_NOUVEAU": 91, // CRUD
	"SRV_FICHIER_ENREGISTRE": 92,
	"SRV_FICHIER_SUPPRIME": 94,
	"SRV_FICHIER_CREATION": 93,
	"SRV_FICHIER_SUIVANT": 98,
	"SRV_FICHIER_PRECEDENT": 99,
	"SRV_FICHIER_PREMIER": 96,
	"SRV_FICHIER_DERNIER": 97,
	"SRV_FICHIER_PRINT": 1796,
	"SRV_FICHIER_PRINT_LISTE": 1797,
	"SRV_FICHIER_COMMENTAIRE": 1795,
	"SRV_FICHIER_PROPRIETE": 1798,
	"SRV_FICHIER_STATISTIQUE": 1823,
	"SRV_FICHIER_WORKFLOW": 1824,
	"SRV_FICHIER_PIECE_JOINT": 1794,
	"SRV_FICHIER_LIAISON": 1825,
	"SRV_SEL_SELECTION": 95, // Obsolete?
	"SRV_SEL_SEL_MOT_CLE": 1100, // Obsolete?
	"SRV_SEL_CHANGE_MOT_CLE": 1793,
	"SRV_EXECUTE": 88,
	"SRV_CHAMP_INACTIF": 34,
	"SRV_ADXEXECUTE": 162,
	"SRV_CALL": 53,
	"SRV_FUNC": 54,
	"SRV_DOCMSO_NEW": 1848,
	"SRV_EXPXLSGRID": 163,
	"SRV_TUNNEL": 129
};

var _srvActionEx = {
	"SUP_FIELD_DRILL": 3,
	"SUP_FIELD_ICOCLICK": 4,
	"SUP_FIELD_CTX": 5,
	"SUP_CELL_EDIT": 6,
	"SUP_CELL_ICOCLICK": 7,
	"SUP_CELL_ICOCLICKFOCUS": 8,
	"SUP_CELL_CHECK": 9, // unused
	"SUP_CELL_CHECKFOCUS": 10, // unused
	"SUP_LIST_PREV": 11,
	"SUP_LIST_NEXT": 12,
	"SUP_LIST_SELLINE": 13,
	"SUP_LIST_FILTER": 14,
	"SUP_WIN_CLOSESTACK": 15,
	"SUP_LIST_SELLINEEX": 16,
	"SUP_FIELD_SELECT": 17,
	"SUP_LINE_DELETE": 18,
	"SUP_LINE_INSERT": 19,
	"SUP_LINES_DELETE": 20,
	"SUP_CELL_FOCUS_DRILL": 21,
	"SUP_CELL_EDIT_DRILL": 22,
	"SUP_CELL_FOCUS_EDIT_DRILL": 23,
	"SUP_CELL_FOCUS_SELECT": 24,
	"SUP_CELL_EDIT_SELECT": 25,
	"SUP_CELL_FOCUS_EDIT_SELECT": 26,
	"SUP_CELL_FOCUS_CTX": 27,
	"SUP_CELL_EDIT_CTX": 28,
	"SUP_CELL_FOCUS_EDIT_CTX": 29,
	"SUP_LIST_PICKING_FOCUS_SELECT": 30, // set focus and select picking one line
	"SUP_LIST_PICKING_SELECT": 31, // select picking one line
	"SUP_LIST_PICKING_FOCUS_UNSELECT": 32, // set focus and unselect picking one line
	"SUP_LIST_PICKING_UNSELECT": 33, // unselect picking one line
	"SUP_CELL_EDIT_TAB": 34, // set edit and tab
	"SUP_LIST_TREESELECTLINE": 35, // set focus and Select leaf recursive hierarchy
	"SUP_LIST_TREECHANGELINE": 36, // set focus and Select leaf hierarchy
	"SUP_LIST_TREEEXPANDLINE": 37, // set focus and expend line hierarchy
	"SUP_EXPXLSGRID": 38, // export excel full grid
	"SUP_LIST_ABANDON": 39, // cancel grid (cancel line + tab)
	"SUP_FIELD_SRVCALL": 40, // focus field + ack_call
	"SUP_CELL_FOCUS_SRVCALL": 41, // focus grid cell + ack_call
	"SUP_FIELD_SRVFUNC": 42, // focus field + ack_func
	"SUP_CELL_FOCUS_SRVFUNC": 43, // focus grid cell + ack_func
	"SUP_WINS_CLOSE": 44 //Close all wins
};
var _proxyAction = {
	"PRINTER_SEL_OPTS": 1,
	"PRINTER_SEL_CANCEL": 2,
	"PRINTER_SEL_OK": 3,
	"PRINTER_OPTS_CANCEL": 4,
	"PRINTER_OPTS_OK": 5,
	"PLUGIN_ILOG_ACK": 10
};

exports.proxyActionStart = 512;
exports.superActionStart = 768;
exports.srvActionStart = exports.proxyActionStart;
exports.localActionStart = 256;

exports.svcFormat = 129;
exports.svcPrint = 130;
exports.svcHelp = 131;

exports.getSrvAction = function(action) {
	return (_srvAction[action] + 1024);
};
exports.getSrvActionEX = function(action) {
	return (_srvActionEx[action] + exports.superActionStart);
};
exports.getProxyAction = function(action) {
	return (_proxyAction[action] + exports.proxyActionStart);
};
exports.getActionByCategorie = function(cat, actionId) {
	var act = null;
	switch (cat) {
		case "proxyaAct":
			act = exports.getProxyAction(actionId);
			break;
		case "superAct":
			act = exports.getSrvActionEX(actionId);
			break;
		default:
			act = exports.getSrvAction(actionId);
			break;
	}
	return act;
};

exports.specialSrvActId = {
	"selTransact": 2844,
	"refresh": 2823,
	"lastRead": 1167,
	"converter": 1169
};

var _kActFocusSelected = ["SRV_CHAMP_SUIVANT", "SRV_CHAMP_INACTIF", "SRV_CHAMP_PRECEDENT", "SRV_FICHIER_NOUVEAU", "SRV_FICHIER_CREATION"];
var _kActFldFocus = ["SRV_GETFOCUS", "SRV_GETFOCUSCELL", "SRV_COMMANDGRID", "SRV_KIB"];
exports.kActionsMagneto = ["SRV_FICHIER_SUIVANT", "SRV_FICHIER_PRECEDENT", "SRV_FICHIER_PREMIER", "SRV_FICHIER_DERNIER"];

function _findSrvActInlist(actionNum, actList) {
	var ret = null,
		i, len;
	for (i = 0, len = actList.length; i < len; i++) {
		if (exports.getSrvAction(actList[i]) == actionNum) {
			ret = actList[i];
			break;
		}
	}
	return ret;
}
exports.isActFocusSelected = function(actionNum) {
	return !!(_findSrvActInlist(actionNum, _kActFocusSelected) != null);
};
exports.isActFldFocus = function(actionNum) {
	return !!(_findSrvActInlist(actionNum, _kActFldFocus) != null);
};
exports.isActMagneto = function(actionNum) {
	return !!(_findSrvActInlist(actionNum, exports.kActionsMagneto) != null);
};
