// syracuse-ui/lib/fusion/core/client/sap/winMenu.js
"use strict";

/*
 Copyright (c)2012 Sage.
 http://wwww.sage.com
 */
/**
 * @fileoverview Fusion page menu management
 */


var helpers = require('syracuse-core/lib/helpers');
var requestActions = require('./srvactions');
var util = require('syracuse-ui/lib/fusion/tools/util');
var syraUtil = require('./util').Syra;
var sapUtil = require('./util').Fusion;
var utilBis = require('syracuse-ui/lib/fusion/tools/utilFdb');

var _menuWithMagneto = true;
var _menuFlat = false;
var _menuIDs = {	
	// "File" menu : common for all types of window
	"FILE_EX": "322",
	// "Print" submenu : common for all types of window
	"SUBMENU_PRINT_CMN": "319",
	// "Help" menu : common for all types of window
	"HELP_EX": "323",
	// "Diagnos" submenu : common for all types of window
	"SUBMENU_DIAGNOS_CMN": "318",
	// "Display" menu : common for all types of window
	"DISPLAY_EX": "324",
	// "Navigation" menu for "Girl" type window
	"NAVIGATION_GIRL": "325",
	// "Window" menu for "Girl" type window
	"WINDOW_GIRL": "326",
	// "Navigation" menu for "Own" type window
	"NAVIGATION_OWN": "320",
	// "Window" menu for "Own" type window
	"WINDOW_OWN": "321", 
	// "File" menu for "Mother" window
	"FILE_MOTHER": "276",
	"NAVIGATION_MOTHER": "308",
	// "Tool" menu for "Mother" window
	"TOOL_MOTHER": "301",
	"SUBMENU_DIAGNOS_MOTHER": "317",
	// "Help" menu for "Mother" window
	"HELP_MOTHER": "312",
	// "Edition" menu for all types of window
	"EDITION_WIN": "278",
	// Misc ID of "edition" menu items....
	"EDITION_UNDO": "289",
	"EDITION_CUT": "290",
	"EDITION_COPY": "291",
	"EDITION_PASTE": "292",
	// "RichEdit" menu for all types of window
	"RICHEDIT_WIN": "283",
	// Miscs menu for V140 window "Girl"
	"FILE_GIRL_V140": "277",
	"TOOL_GIRL_V140": "299",
	"DISPLAY_GIRL_V140": "279",
	"HELP_V140": "280",
	// Miscs menu for V140 window "Own"
	"FILE_OWN_V140": "293",
	"DISPLAY_OWN_V140": "295",
	"SELECTION_OWN_V140": "311", 
	"TOOL_OWN_V40": "300",
	// Miscs menu for Editor windows
	"FILE_EDITORL4G": "296",
	"TOOL_EDITORL4G": "298",
	"FILE_EDITORRTF": "302",
	"EDITION_EDITOR": "297",
	"FILE_EDITORGRA": "303",
	"EDITION_EDITORGRA": "304",
	"TOOL_EDITORGRA": "305",
	"HELP_MINI_EDITORS": "313",
	// Obsolet menus
	"SELECTION_GIRL_OBSOLET": "310",
	"APPLICATIVE_OBSOLET": "327",
	// Not Used ID
	"SEPARATOR": "275",
	"DUMMY": "294",
	// Special
	"ROOT_SUPERVISOR": "307",
	"BTN_WIN" : "997",
	"MAGNETO_GIRL" : "998",
	"MAGNETO_OWN" : "999"
};

function _addMenus (menusObj, tabMenusFather, menus, options){
	if(menusObj && menus) {
	   	_.each(menus, function(menu) {
	    	_addMenu(menusObj, tabMenusFather, menu, options);
	    });
	}
}

function _addMenu(menusObj, tabMenusFather, menu, options) {
	var father = tabMenusFather && tabMenusFather[menu.fid] ? tabMenusFather[menu.fid] : menusObj[menu.fid];
	if(!father) {
		if(options && options.logger) {
			options.logger.debug({"severety": "warn"}, "'winMenu._addMenu' : attemp to add an item menu to an unknown menu id (" + menu.fid +  ")");
		}
		return;
	}
	if(!father.items) {
		father.items = {};
	}
	father.items[menu.id] = menu;
	if(menu.typ === 0) {
		father.items[menu.id].items = {};
		if(tabMenusFather) {
			tabMenusFather[menu.id] = father.items[menu.id];
		}
	}
}

function _releaseMenusObj (menusObj){
	_.each(menusObj, function(menu, key){
		if (menu.items) {
			_releaseMenusObj(menu.items);
		}
		delete this[key];
	}, menusObj);
}

function MenuManagement(){
}

exports.MenuManagement = helpers.defineClass(MenuManagement, null, {
    initialize: function(session, commonMenus){
    	var timer = utilBis.getTimer(true);
    	this._session = session;
		this._idMnuCpt = 512;
		this._mnuWithMagneto = _menuFlat ? false : _menuWithMagneto;
    	this.winMenuColl = {};
    	this._tabMenusFather = {};
    	this._coreMenus = this._getCoreMenuIDs();
    	_addMenus(this._coreMenus, this._tabMenusFather, commonMenus, {"logger": this._session.logger});
   		this._addInternalCoreMenu();
        this._session.logger.timer(null, "Load core menus : total [" + timer.elapsedMs() + "]");
     },

    _addInternalCoreMenu: function() {
    	var mnuId;
    	/*Add items to root menus*/
    	// Items of "DISPLAY_EX"
    	_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(_menuIDs.DISPLAY_EX, this._idMnuCpt++, null, "9", 1, 51, null, "SRV_FICHIER_STATISTIQUE"));
    	// Items of "NAVIGATION_MOTHER"
   		_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(_menuIDs.NAVIGATION_MOTHER, this._idMnuCpt++, null, "10", 1, 20, null, "SRV_ADXEXECUTE"));
   		// Items of "SUBMENU_DIAGNOS_MOTHER"
   		_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(_menuIDs.SUBMENU_DIAGNOS_MOTHER, this._idMnuCpt++, null, "11", 1, 10, null, "SRV_CALCULATRICE"));
   		_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(_menuIDs.SUBMENU_DIAGNOS_MOTHER, this._idMnuCpt++, null, "12", 1, 20, null, "SRV_DEBUGGER"));
    	// Items of "WINDOW_OWN"
   		_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(_menuIDs.WINDOW_OWN, this._idMnuCpt++, null, "13", 1, 11, null, null, true));
    	// Items of "WINDOW_GIRL"
   		_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(_menuIDs.WINDOW_GIRL, this._idMnuCpt++, null, "13", 1, 11, null, null, true));
    	// Items of "SUBMENU_DIAGNOS_CMN"
  		_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(_menuIDs.SUBMENU_DIAGNOS_CMN, this._idMnuCpt++, null, "11", 1, 11, null, "SRV_CALCULATRICE"));
 		_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(_menuIDs.SUBMENU_DIAGNOS_CMN, this._idMnuCpt++, null, "12", 1, 12, null, "SRV_DEBUGGER"));
   		// Items of "FILE_EX"
  		_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(_menuIDs.FILE_EX, this._idMnuCpt++, null, "14", 1, 5, null, "SRV_FICHIER_NOUVEAU"));
  		_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(_menuIDs.FILE_EX, this._idMnuCpt++, null, "15", 1, 20, null, "SRV_SEL_CHANGE_MOT_CLE"));		   		
  		_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(_menuIDs.FILE_EX, this._idMnuCpt++, null, "0", 2, 40, null, null));
  		_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(_menuIDs.FILE_EX, this._idMnuCpt++, null, "16", 1, 41, null, "SRV_FICHIER_PIECE_JOINT"));		   		
  		_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(_menuIDs.FILE_EX, this._idMnuCpt++, null, "17", 1, 42, null, "SRV_FICHIER_COMMENTAIRE"));		   		
   		_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(_menuIDs.FILE_EX, this._idMnuCpt++, null, "0", 2, 60, null, null));		   		
   		_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(_menuIDs.FILE_EX, this._idMnuCpt++, null, "0", 2, 70, null, null));		   		
  		_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(_menuIDs.FILE_EX, this._idMnuCpt++, null, "20", 1, 71, null, "SRV_FICHIER_WORKFLOW"));		   		
  		_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(_menuIDs.FILE_EX, this._idMnuCpt++, null, "21", 1, 72, null, "SRV_FICHIER_PROPRIETE"));		   		
  		// Items of HELP_EX
		_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(_menuIDs.HELP_EX, this._idMnuCpt++, null, "0", 2, 50, null, null));
   		// Items of "SUBMENU_PRINT_CMN" 
  		_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(_menuIDs.SUBMENU_PRINT_CMN, this._idMnuCpt++, null, "18", 1, 10, null, "SRV_FICHIER_PRINT"));		   		
  		_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(_menuIDs.SUBMENU_PRINT_CMN, this._idMnuCpt++, null, "19", 1, 20, null, "SRV_FICHIER_PRINT_LISTE"));		   		
   		// Items of "NAVIGATION_OWN"
   		mnuId = !this._mnuWithMagneto ? _menuIDs.NAVIGATION_OWN : _menuIDs.MAGNETO_OWN;
  		_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(mnuId, this._idMnuCpt++, null, "22", 1, 11, null, "SRV_FICHIER_PREMIER", false, null, "$first"));
  		_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(mnuId, this._idMnuCpt++, null, "23", 1, 12, null, "SRV_FICHIER_PRECEDENT", false, null, "$previous"));		   		
  		_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(mnuId, this._idMnuCpt++, null, "24", 1, 13, null, "SRV_FICHIER_SUIVANT", false, null, "$next"));		   		
  		_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(mnuId, this._idMnuCpt++, null, "25", 1, 15, null, "SRV_FICHIER_DERNIER", false, null, "$last"));		   		
  		if(!this._mnuWithMagneto) {
   			_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(_menuIDs.NAVIGATION_OWN, this._idMnuCpt++, null, "0", 2, 20, null, null));	   		
   		}
  		_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(_menuIDs.NAVIGATION_OWN, this._idMnuCpt++, null, "20", 1, 21, null, "SRV_ADXEXECUTE"));		   		
 		_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(_menuIDs.NAVIGATION_OWN, this._idMnuCpt++, null, "26", 1, 55, null, "SRV_FICHIER_LIAISON"));		   		
   		// Items of "NAVIGATION_GIRL" 
   		mnuId = !this._mnuWithMagneto ? _menuIDs.NAVIGATION_GIRL : _menuIDs.MAGNETO_GIRL;
  		_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(mnuId, this._idMnuCpt++, null, "22", 1, 11, null, "SRV_FICHIER_PREMIER", false, null, "$first"));		   		
  		_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(mnuId, this._idMnuCpt++, null, "23", 1, 12, null, "SRV_FICHIER_PRECEDENT", false, null, "$previous"));		   		
  		_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(mnuId, this._idMnuCpt++, null, "24", 1, 13, null, "SRV_FICHIER_SUIVANT", false, null, "$next"));		   		
  		_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(mnuId, this._idMnuCpt++, null, "25", 1, 15, null, "SRV_FICHIER_DERNIER", false, null, "$last"));		   		
  		if(!this._mnuWithMagneto) {
   			_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(_menuIDs.NAVIGATION_GIRL, this._idMnuCpt++, null, "0", 2, 20, null, null));	   		
   		}
  		_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(_menuIDs.NAVIGATION_GIRL, this._idMnuCpt++, null, "20", 1, 21, null, "SRV_ADXEXECUTE"));		   		
 		_addMenu(this._coreMenus, this._tabMenusFather, this._getMenuItem(_menuIDs.NAVIGATION_GIRL, this._idMnuCpt++, null, "26", 1, 55, null, "SRV_FICHIER_LIAISON"));		   		

 		// Now, attach core submenus to root menus*/
 		this._addCoreSubMenuToRoot(_menuIDs.FILE_EX, _menuIDs.SUBMENU_PRINT_CMN, 65);
  		this._addCoreSubMenuToRoot(_menuIDs.HELP_EX, _menuIDs.SUBMENU_DIAGNOS_CMN, 51);
  		this._addCoreSubMenuToRoot(_menuIDs.TOOL_MOTHER, _menuIDs.SUBMENU_DIAGNOS_MOTHER, 50);
	},

	_addCoreSubMenuToRoot: function(fid, id, rank){
		if(!this._coreMenus[fid].items) {
			this._coreMenus[fid].items = {};
		}
 		this._coreMenus[fid].items[id] = this._coreMenus[id];
 		this._coreMenus[fid].items[id].rnk = rank;	   		
 		this._coreMenus[fid].items[id].typ = 0;
	},

	_getMenuItem: function (fid, id, txt, txtNum, typ, rnk, act, actTxt, inactive, opts, linkId){
    	return {"fid": fid, "id": id.toString(), "txt": txt || this._session.getlabel("text", txtNum, txtNum), "typ": typ, "rnk": rnk, "act": act || requestActions.getSrvAction(actTxt), "inactive": inactive, "linkId": linkId, "opts": (opts ? util.duplicateObj(opts) : undefined)};
    },

	addMnuBarToProto: function(winId, winFacet, winMenusDef, winProto){
 		var i, len, linksArray, layout, timer = utilBis.getTimer(true), bar = this._getWinMenuBar(winId, winFacet, winMenusDef, winProto, _menuFlat);
		if(!_menuFlat) {
			// Set links
			if(winProto.$links) {
				linksArray = Object.keys(winProto.$links);
				for(i = 0, len = linksArray.length; i < len; i++ ) {
					bar.$links[linksArray[i]] = util.duplicateObj(winProto.$links[linksArray[i]]);
				}
				delete winProto.$links;
			}
			winProto.$links = bar.$links;
			// Set menu layout, consider that items set in proto are memebers of "ROOT_SUPERVISOR", so the last group in the bar
			syraUtil.appendProto(winProto, {"$article": {"$menus": {"$layout": {"$items" : []}}}});
			layout = winProto.$article.$menus.$layout;
			if(layout.$items.length == 0) {
				layout.$items = bar.$menus.$items;
			}
			else {
				Array.prototype.unshift.apply(layout.$items, bar.$menus.$items);
			}
		}
		else {
			sapUtil.appendMnus2Syra(winProto, bar, true);
		}
        this._session.logger.timer(null, "Add Menu Bar Definition : total [" + timer.elapsedMs() + "]");
		//this._session.logger.trace({"toJson": true}, "Menu Bar Syracuse : ", bar);
	},

	releaseWinMenu: function(winId){
		var winMenu;
		if((winMenu = this.winMenuColl[winId])) {
			if(winMenu.winProto) {
		 		delete winMenu.winProto.$article.$menus;
		 		delete winMenu.winProto.$links;
		 	}
			winMenu.dispose();
			delete this.winMenuColl[winId];
		}	
	},

	_getWinMenuBar: function(winId, winFacet, winMenusDef, winProto, flat){
		if(this.winMenuColl[winId]) {
			this.releaseWinMenu(winId);
		}
		(this.winMenuColl[winId] = new WinMenu()).initialize(winMenusDef, this._coreMenus, winProto, winFacet, flat, this._mnuWithMagneto, this._session);
		this.winMenuColl[winId].buildMenus();
		//this._session.logger.trace({"toJson": true}, "Menu Bar JSON : ", this.winMenuColl[winId]._menuBarTree);
		return this.winMenuColl[winId].getMenuBar();
	},
	
    dispose: function(){
  		_.each(this.winMenuColl, function(ptr, key){this.releaseWinMenu(key);}, this);
  		_.each(this._tabMenusFather, function(ptr, key){delete this[key];}, this._tabMenusFather);
   		_releaseMenusObj(this._coreMenus);
    	if(this._session) {
    		delete this._session;
    	}
    },

   	_getCoreMenuIDs: function() {
  		var ids = {};
		// Common for all types of window
		ids[_menuIDs.FILE_EX] = {"tag":"FILE_EX", "txt": this._session.getlabel("text", 2, "File"), "typ": 0, "items": {}};
		ids[_menuIDs.SUBMENU_PRINT_CMN] = {"tag":"SUBMENU_PRINT_CMN", "txt": this._session.getlabel("text", 3, "Print"), "typ": 0};
		ids[_menuIDs.HELP_EX] = {"tag":"HELP_EX", "txt": this._session.getlabel("text", 4, "Help"), "typ": 0};
		ids[_menuIDs.SUBMENU_DIAGNOS_CMN] = {"tag":"SUBMENU_DIAGNOS_CMN", "txt": this._session.getlabel("text", 5, "Diagnosis"), "typ": 0};
		ids[_menuIDs.DISPLAY_EX] = {"tag":"DISPLAY_EX", "txt": this._session.getlabel("text", 6, "Display"), "typ": 0};

		// "Girl" menus
		ids[_menuIDs.NAVIGATION_GIRL] = {"tag":"NAVIGATION_GIRL", "txt": this._session.getlabel("text", 7, "Navigation"), "typ": 0};
		ids[_menuIDs.WINDOW_GIRL] = {"tag":"WINDOW_GIRL", "txt": this._session.getlabel("text", 8, "Window"), "typ": 0};

		// "Own" menus
		ids[_menuIDs.NAVIGATION_OWN] = {"tag":"NAVIGATION_OWN", "txt": this._session.getlabel("text", 7, "Navigation"), "typ":0};
		ids[_menuIDs.WINDOW_OWN] = {"tag":"WINDOW_OWN", "txt": this._session.getlabel("text", 8, "Window"), "typ":0};

		// "Special" magneto menus
		ids[_menuIDs.MAGNETO_GIRL] = {"tag": "MAGNETO_GIRL"}; 
		ids[_menuIDs.MAGNETO_OWN] = {"tag": "MAGNETO_OWN"}; 

		// "Mother" menus
		ids[_menuIDs.FILE_MOTHER] = {"tag":"FILE_MOTHER", "txt": this._session.getlabel("text", 2, "File"), "typ": 0};
		ids[_menuIDs.NAVIGATION_MOTHER] = {"tag":"NAVIGATION_MOTHER", "txt": this._session.getlabel("text", 7, "Navigation"), "typ": 0};
		ids[_menuIDs.TOOL_MOTHER] = {"tag":"TOOL_MOTHER"};
		ids[_menuIDs.SUBMENU_DIAGNOS_MOTHER] = {"tag":"SUBMENU_DIAGNOS_MOTHER", "txt": this._session.getlabel("text", 5, "Diagnosis"), "typ": 0};
		ids[_menuIDs.HELP_MOTHER] = {"tag":"HELP_MOTHER", "txt": this._session.getlabel("text", 4, "Help"), "typ": 0};

		// "Edition" menu for all types of window
		ids[_menuIDs.EDITION_WIN] = {"tag":"EDITION_WIN"};
		// Misc ID of "edition" menu items....
		ids[_menuIDs.EDITION_UNDO] = {"tag":"EDITION_UNDO"};
		ids[_menuIDs.EDITION_CUT] = {"tag":"EDITION_CUT"};
		ids[_menuIDs.EDITION_COPY] = {"tag":"EDITION_COPY"};
		ids[_menuIDs.EDITION_PASTE] = {"tag":"EDITION_PASTE"};
		// "RichEdit" menu for all types of window
		ids[_menuIDs.RICHEDIT_WIN] = {"tag":"RICHEDIT_WIN"};

		// Miscs menu for V140 window "Girl"
		ids[_menuIDs.FILE_GIRL_V140] = {"tag":"FILE_GIRL_V140"};
		ids[_menuIDs.TOOL_GIRL_V140] = {"tag":"TOOL_GIRL_V140"};
		ids[_menuIDs.DISPLAY_GIRL_V140] = {"tag":"DISPLAY_GIRL_V140"};
		ids[_menuIDs.HELP_V140] = {"tag":"HELP_V140"};

		// Miscs menu for V140 window "Own"
		ids[_menuIDs.FILE_OWN_V140] = {"tag":"FILE_OWN_V140"};
		ids[_menuIDs.DISPLAY_OWN_V140] = {"tag":"DISPLAY_OWN_V140"};
		ids[_menuIDs.SELECTION_OWN_V140] = {"tag":"SELECTION_OWN_V140"};
		ids[_menuIDs.TOOL_OWN_V40] = {"tag":"TOOL_OWN_V40"};

		// Miscs menu for Editor windows
		ids[_menuIDs.FILE_EDITORL4G] = {"tag":"FILE_EDITORL4G"};
		ids[_menuIDs.TOOL_EDITORL4G] = {"tag":"TOOL_EDITORL4G"};
		ids[_menuIDs.FILE_EDITORRTF] = {"tag":"FILE_EDITORRTF"};
		ids[_menuIDs.EDITION_EDITOR] = {"tag":"EDITION_EDITOR"};
		ids[_menuIDs.FILE_EDITORGRA] = {"tag":"FILE_EDITORGRA"};
		ids[_menuIDs.EDITION_EDITORGRA] = {"tag":"EDITION_EDITORGRA"};
		ids[_menuIDs.TOOL_EDITORGRA] = {"tag":"TOOL_EDITORGRA"};
		ids[_menuIDs.HELP_MINI_EDITORS] = {"tag":"HELP_MINI_EDITORS"};

		// Obsolet menus
		ids[_menuIDs.SELECTION_GIRL_OBSOLET] = {"tag":"SELECTION_GIRL_OBSOLET"};
		ids[_menuIDs.APPLICATIVE_OBSOLET] = {"tag":"APPLICATIVE_OBSOLET"};
		
		// Not Used ID
		ids[_menuIDs.SEPARATOR] = {"tag":"SEPARATOR"};
		ids[_menuIDs.DUMMY] = {"tag":"DUMMY"};
		return ids;
	},

   	_getCoreMenuCtxIDs: function() {
		return {		
			"309": {"tag":"QUICKHELP"}, "281": {"tag":"CTX_GRID"}, "282": {"tag":"CTX_LISTVIEW"}, "284": {"tag":"CTX_RICHEDIT"}, "285": {"tag":"CTX_EDIT"}, "287": {"tag":"CTX_CHECK"},
			"286": {"tag":"CTX_COMBO"}, "288": {"tag":"CTX_RADIO"}, "306": {"tag":"CTX_TREEVIEW"}, "314": {"tag":"CTX_IMAGE"}, "315": {"tag":"CTX_BUTTON"}, "316": {"tag":"CTX_XBLOCK"} 
		};
	}
});

function WinMenu(){
}

WinMenu = helpers.defineClass(WinMenu, null, {
  	initialize: function(winMenus, coreMenus, winProto, winFacet, flat, showMagneto, session){
  		this._winFacet = winFacet;
  		this._flat = flat;
  		this._showMagneto = showMagneto;
  		this._coreMenus = coreMenus;
  		this._session = session;
    	this._tabMenusFather = {};
    	this._menuBarTree = {"items": {}};
    	this._menuBarObject = null;
 		this._menus = this._getSpecialMenuIDs();
 		if(winMenus && winMenus.length > 0) {
    		_addMenus(this._menus, this._tabMenusFather, winMenus, {"logger": this._session.logger});
    	}
    	if(winProto.$actions) {
    		this._addWinProtoAction(winProto.$actions, winProto.$localization);
	        delete winProto.$actions;
    	}
    	this.winProto = winProto;
   	},

   	getMenuBar: function() {
   		if(!this._menuBarObject) {
			this._menuBarObject = this._flat ? this.buildFlatMenusObj() : this.buildTreeMenusObj();
		}
		return this._menuBarObject;
   	}, 
 
  	_addWinProtoAction: function(actions, localization){
   		var len, actsdArray, i, actItm;
  		actsdArray = Object.keys(actions);
        for (i = 0, len = actsdArray.length; i < len; i++) {
        	actItm = actions[actsdArray[i]];
        	_addMenu(this._menus, null, {"id": actsdArray[i], "fid": _menuIDs.BTN_WIN, "rnk": i, typ: "1", "act": actItm.$act, "txt": syraUtil.getLocalization(actItm.$title, localization)}, {"logger": this._session.logger});
        }
        this._hasActWin = len > 0 ? true : false;
  	},
  
   	buildMenus: function(){
   		var map, appMemus, self = this;
   		// Map standards menus by group...
   		if(!this._winFacet || this._winFacet === syraUtil.pageFacet.edit) {
   			map = [{"id": _menuIDs.FILE_EX, "rnk": 5}, {"id": _menuIDs.DISPLAY_EX, "rnk": 10}, {"id": _menuIDs.NAVIGATION_GIRL, "rnk": 15}, {"id": _menuIDs.WINDOW_GIRL, "rnk": 97}, {"id": _menuIDs.HELP_EX, "rnk": 98}];
    	}
   		else {
 			map = [{"id": _menuIDs.FILE_EX, "rnk": 5}, {"id": _menuIDs.DISPLAY_EX, "rnk": 10}, {"id": _menuIDs.NAVIGATION_OWN, "rnk": 15}, {"id": _menuIDs.WINDOW_OWN, "rnk": 97}, {"id": menuIDs.HELP_EX, "rnk": 98}];
   		}
   		_.each(map, function(menu) {
   			this[menu.id] = self._coreMenus[menu.id];
   			this[menu.id].rnk = menu.rnk;
   		}, this._menuBarTree.items);
   		// ... add "actions" menus...
   		if(this._hasActWin) {
   			this._menuBarTree.items[_menuIDs.BTN_WIN] = this._menus[_menuIDs.BTN_WIN];
   		}
   		// ... then add applicative menu groups
   		if((appMemus = this._menus[_menuIDs.ROOT_SUPERVISOR].items)) {
	    	_.each(appMemus, function(menu) {
	   			this[menu.id] = appMemus[menu.id];
	   			this[menu.id].rnk = menu.rnk;
	   		}, this._menuBarTree.items);
	    }
   	},

   	buildTreeMenusObj: function(){
   		var $menus = {"$items": []}, $links = {}, magnetoMnuId, i, len, magnetoArray, magetoItem;
		function walk(srcItems, trgtItems) {
    		var len, srcItemsArray, i, act, style;
			srcItemsArray = (Object.keys(srcItems)).sort(function(left, right) {
				if (srcItems[left].rnk > srcItems[right].rnk || srcItems[left].rnk === void 0) {
					return 1;
				}
				if (srcItems[left].rnk < srcItems[right].rnk || srcItems[right].rnk === void 0) {
					return -1;	
				}
				else {
					return 0;
				}
			});
			for (i = 0, len = srcItemsArray.length; i < len; i++) {
	    		if (!srcItems[srcItemsArray[i]].inactive && srcItems[srcItemsArray[i]].typ != 2) {
	   				style = srcItems[srcItemsArray[i]].opts ? srcItems[srcItemsArray[i]].opts.sty : undefined;
	    			if(srcItems[srcItemsArray[i]].typ === 0) {
	    				trgtItems.push({"$title": srcItems[srcItemsArray[i]].txt, "$style": style, "$layout": {"$items": []}});
	    				walk(srcItems[srcItemsArray[i]].items, trgtItems[trgtItems.length - 1].$layout.$items);
	    			}
	    			else if ((act = srcItems[srcItemsArray[i]].act)){
	    				$links[act] = {"$act": act, "$title": srcItems[srcItemsArray[i]].txt};
	    				trgtItems.push({"$bind": act, "$style": style, "rnk": srcItems[srcItemsArray[i]].rnk});
	    			}
	    		}
			}
   		}
   		// Walk accross our menu tree
   		walk(this._menuBarTree.items, $menus.$items);
   		// Add magneto menu... if necessary
   		if(this._showMagneto) {
   			magnetoMnuId = !this._winFacet || this._winFacet === syraUtil.pageFacet.edit ? _menuIDs.MAGNETO_GIRL : _menuIDs.MAGNETO_OWN;
   			magnetoArray = this._coreMenus[magnetoMnuId] && this._coreMenus[magnetoMnuId].items ? Object.keys(this._coreMenus[magnetoMnuId].items) : [];
   			for(i = 0, len = magnetoArray.length; i < len; i++) {
   				magetoItem = this._coreMenus[magnetoMnuId].items[magnetoArray[i]];
   				$links[magetoItem.linkId] = {"$title" : magetoItem.txt, "$act":magetoItem.act};
   			}
   		}
 		return {"$menus": $menus, "$links" : $links};
   	},

  	buildFlatMenusObj: function(){
   		var menus = [];
   		function walk(items) {
	    	_.each(items, function(menu) {
	    		if (!menu.inactive && menu.typ != 2) {
	    			if(menu.typ === 0) {
	    				walk(menu.items);
	    			}
	    			else {
	    				menus.push(menu);
	    			}
	    		}
	   		});
   		}
   		walk(this._menuBarTree.items);
 		return menus;
   	},
   
   	dispose: function(){
	 	if(this.winProto) {
	 		delete this.winProto;
	 	}
 	 	_.each(this._tabMenusFather, function(ptr, key){delete this[key];}, this._tabMenusFather);
	 	_.each(this._menuBarTree.items, function(ptr, key){delete this[key];}, this._menuBarTree.items);
	 	_releaseMenusObj(this._menus);
	 	if(this._menuBarObject) {
		 	if (this._flat){
			 	_.each(this._menuBarObject, function(ptr, idx){delete this[idx]}, this._menuBarObject);
		 	}
		 	else {
		 		delete this._menuBarObject.$menus;
		 		delete this._menuBarObject.$links;
		 	}
	 	}
    	if(this._coreMenus) {
    		delete this._coreMenus;
    	}
    	if(this._session) {
    		delete this._session;
    	}
    },

  	_getSpecialMenuIDs: function() {
  		var ids = {};
  		ids[_menuIDs.ROOT_SUPERVISOR] = {"tag": "ROOT_SUPERVISOR"}; 
		ids[_menuIDs.BTN_WIN] = {"tag": "BTN_WIN", "typ": 0, "rnk": 1, "opts": {"sty": "main"}};
		return ids;
	}
});
