// syracuse-ui/lib/desktop/field/choice/choiceField.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var ChoiceField = require('syracuse-ui/lib/common/field/choiceField').ChoiceField;
var ChoiceFieldAuthoring = require('./choiceFieldAuthoring').ChoiceFieldAuthoring;

var _formatBuilders = {
    $combo: require('./comboFormat').ComboFormat,
    $radios: require('./radiosFormat').RadiosFormat
};

function DesktopChoiceField(){
}

exports.DesktopChoiceField = helpers.defineClass(DesktopChoiceField, ChoiceField, {
    applyDesignMetaData: function(metadata, onAuthoring){
        ChoiceField.prototype.applyDesignMetaData.call(this, metadata, onAuthoring);
        var updated;
        if (metadata.$format && metadata.$format != this.$item.$format) {
            if (metadata.$choiceLayout !== undefined) {
                this.$item.$choiceLayout = metadata.$choiceLayout;
            }
            if (this.builder) {
                document.controller.disposeObject(this.builder);
                delete this.builder;
            }
            this.$item.$format = metadata.$format;
            this.$$dataValue.empty();
            this.render();
            updated = true;
        }
        if (metadata.$choiceLayout !== undefined && metadata.$choiceLayout != this.$item.$choiceLayout) {
            this.$item.$choiceLayout = metadata.$choiceLayout;
            this.builder.renderLayout();
            updated = true;
        }
        if (updated) {
            this.setState(this.$field);
            if (this.currentValue != null) {
                this.setDataValue(this.currentValue)
            }
        }
    },
    setDataValue: function(value){
        this.builder.setDataValue(value);
    },
    getAuthoringWidget: function(){
        return new ChoiceFieldAuthoring();
    },
    getInputValue: function(){
        return this.builder.getInputValue();
    },
    render: function(){
        this.$item.$format = this.$item.$format || this.$field.$format;
        if (this.$item.$isFilterMode || this.$item.$format == undefined && (this.$field.$value.$constraints.$enum.length > (this.$item.$radiosMax || 4))) {
            this.$item.$format = "$combo";
        }
        this.builder = new _formatBuilders[this.$item.$format = (this.$item.$format || "$radios")]();
        this.builder.field = this;
        this.builder._choices = [];
        this.$item.$choiceLayout = this.$item.$choiceLayout || this.builder.$defaultChoiceLayout;
        if (this.builder.field.$item.$iconMode) {
            this.builder.$iconPath = document.site.$item.$iconPath + (this.builder.field.$item.$iconPath || "");
        }
        this.builder.render();
    },
    onInputKeyup: function($$input, event){
        return this.builder.onInputKeyup($$input, event);
    },
    setState: function(state){
        ChoiceField.prototype.setState.call(this, state);
        if (state.$isDisabled !== undefined) {
            this.$$dataValue.find("input").attr("disabled", state.$isDisabled);
        }
        if ((state.$isReadOnly !== undefined) && (this.$item.$format == "$radios")) {
            if (this.$isReadOnly) {
                this.$$dataValue.find("input").attr("disabled", true);
            }
            else {
                if (!this.$isDisabled) {
                    this.$$dataValue.find("input").attr("disabled", false);
                }
            }
        }
    },
    _doClickPicker: function(){
        if (this.builder.doClickPicker) {
            this.builder.doClickPicker();
        }
    },
    releaseMode: function(onDispose){
        if (this.builder) {
            this.builder.releaseMode(onDispose);
        }
        delete this.builder;
        ChoiceField.prototype.releaseMode.call(this, onDispose);
    }
});
