// syracuse-ui/lib/site/url/bookmarks/bookmarksManager.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var BookmarksPage = require("./bookmarksPage").BookmarksPage;
var _dialog;

function _defineMenuItemLinks() {
	return {
		$select: {
			$title: "Add",
			$variants: {
				menuItem: {
					$title: document.site.localize.bookmarksSelectMenu,
					$url: "/sdata/syracuse/collaboration/syracuse/menuItems?representation=menuItem.$select"
				}
			}
		},
		addSeparator: {
			$title: "Add",
			$variants: {
				menuItem: {
					$title: document.site.localize.bookmarksAddSeparator
				}
			}
		}

	};
}

function _defineMenuItem() {
	return {
		$type: "application/x-reference",
		$capabilities: "reorder,delete",
		$item: {
			$value: "{title}",
			$key: "{$uuid}",
			$properties: {
				title: {
					$type: "application/x-string"
				}
			}
		}
	};
}

exports.open = function(bar) {
	var pageSlot = document.createElement("div");
	pageSlot.className = "s-bookmarks-viewer";
	document.site.layoutSlot.appendChild(pageSlot);
	_dialog = document.site.dialogManager.openModal(document.site, {
		$itemPage: {
			$pageCategoryClass: BookmarksPage,
			layoutSlot: pageSlot,
			$category: "page",
			$facet: "$bookmarks",
			$representation: {
				$prototype: {
					$type: "application/json",
					$title: document.site.localize.bookmarksTitle,
					$properties: {
						items: {
							$type: "application/x-array",
							$capabilities: "reorder",
							$links: _defineMenuItemLinks(),
							$item: {
								$type: "application/x-variant",
								$variants: {
									menuItem: _defineMenuItem(),
									menuBlock: {
										$type: "application/x-object",
										$capabilities: "append,reorder,delete",
										$item: {
											$value: "{title}",
											$key: "{$uuid}",
											$id: "menuSubblock",
											$properties: {
												title: {
													$type: "application/x-string"
												},
												items: {
													$type: "application/x-array",
													$capabilities: "reorder",
													$links: _defineMenuItemLinks(),
													$item: {
														$type: "application/x-variant",
														$variants: {
															menuItem: _defineMenuItem(),
															menuBlock: {
																$type: "application/x-object",
																$capabilities: "append,reorder,delete",
																$item: {
																	$prototype: "#menuSubblock",
																	$type: "application/x-pointer"
																}
															}
														}

													}

												}
											}
										}
									}
								}
							}
						}
					}
				},
				$article: {
					$layout: {
						$items: [{
							$category: "section",
							$layout: {
								$items: [{
									$category: "block",
									$layout: {
										$items: [{
											$isEditMode: true,
											$bind: "items",
											$skin: "s-bookmarks",
											$layout: {
												$items: [{
													$bind: "menuItem"
												}]
											}
										}]
									}
								}]
							}
						}]
					}
				}
			}
		}
	});
	exports.refresh(_dialog._content.bar = bar);
};

exports.refresh = function(bar) {
	if (_dialog) {
		_dialog._content.applyChange({
			items: helpers.object.clone(bar.dataset.items, true)
		});
	}
};
exports.dispose = function() {
	if (_dialog) {
		_dialog.dispose();
	}
	_dialog = null;
};
