// syracuse-ui/lib/page/landing/master/landingPageList.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var LandingPageRecord = require("./landingPageRecord").LandingPageRecord;
var MultiList = require("syracuse-ui/lib/field/array/multiList").MultiList;

function LandingPageList() {

}

exports.LandingPageList = helpers.defineClass(LandingPageList, MultiList, {
	loadBox: function(initData) {
		this.isNoPageUniqueSelection = true;
		this.emptyMessage = syra_local.ldpNoPage;
		this.$confirmDelete = syra_local.ldpConfirmDelete;
		this.$isEditMode = false;
		this.$isQuickDesignerDisabled = true;
		this.$itemRecordMenuSkin = "s-mn-ldp-record-link";
		this.$itemMenuSkin = "s-mn-ldp-link";
		this.$iconPath = "page/s-ldp-";
		this.$item.$isPagerHidden = true;
		this.$item.$selectMode = "row";
		MultiList.prototype.loadBox.call(this, initData);
	},
	showRecordSelection: function(record, selected) {
		MultiList.prototype.showRecordSelection.call(this, record, selected);
		var page = record.page;
		if (selected) {
			if (page.selectedPageRecord && record.list != page.selectedPageRecord.list) {
				page.selectedPageRecord.list.selector.selectRecord(page.selectedPageRecord.dataset.$uuid, false);
			}
			page.selectedPageRecord = record;
			record.loadVignettesInChildPage(record.dataset.vignettes);
			page.childSlot.appendChild(record.childPage.domItem);
			page.preferences.$selected = record.dataset.$uuid;
			if (page.isPageLoaded) {
				page.savePreferences();
			}
			page.setTitle(page.$item.$title = record.childPage.getTitle());
			syra_site.siteFunctions.refreshMainPageSecurity(record.childPage);
			record.childPage.ensurePageVisibility();
		} else {
			if (page.selectedPageRecord == record) {
				page.selectedPageRecord = null;
			}
			if (record.childPage) {
				syra_site.dom.removeChild(record.childPage.domItem);
			}
		}
	},
	saveReorder: function(reorderRecord, targetRecord, isAfter) {
		var self = this;
		if (!self.isRoleList()) {
			self.page.preferences.pages = [];
			var children = this.body.children;
			for (var ii = 0, jj = children.length; ii < jj; ii++) {
				self.page.preferences.pages.push(children[ii].syraRecord);
			}
			if (reorderRecord) {
				var reorderIndex = self.page.preferences.pages.indexOf(reorderRecord.$uuid);
				var targetIndex = self.page.preferences.pages.indexOf(targetRecord.$uuid);
				self.page.preferences.pages.splice(reorderIndex, 1);
				if (targetIndex > reorderIndex) {
					reorderIndex = isAfter ? targetIndex : (targetIndex - 2);
				} else {
					reorderIndex = isAfter ? targetIndex + 1 : (targetIndex);
				}
				reorderIndex = Math.max(reorderIndex, 0);
				self.page.preferences.pages.splice(reorderIndex, 0, reorderRecord.$uuid);
			}
			self.page.savePreferences();
			self.applyOrderPreferences();
		}
		return false;
	},
	applyOrderPreferences: function() {
		if (!this.isRoleList() && this.page.preferences.pages) {
			if (this.records && this.records.length) {
				var orders = [];
				for (var ii = 0, jj = this.records.length; ii < jj; ii++) {
					orders.push(this.records[ii].$uuid);
				}
				for (var ii = 0, jj = this.page.preferences.pages.length; ii < jj; ii++) {
					var $uuid = this.page.preferences.pages[ii];
					var record = this.recordsMap[$uuid];
					if (record) {
						record.domItem.parentNode.appendChild(record.domItem);
						orders.splice(orders.indexOf($uuid), 1);
					}
				}
				for (var ii = 0, jj = orders.length; ii < jj; ii++) {
					var record = this.recordsMap[orders[ii]];
					record.domItem.parentNode.appendChild(record.domItem);
				}
			}
		}
	},
	isRoleList: function() {
		return this.$item.$bind == "rolePages";
	},
	drawBox: function() {
		MultiList.prototype.drawBox.call(this);
		this.body = this._core;
		this._dataValue.style.overflow = "auto";
		this.$skin = this.$item.$skin || "s-cards";
		this.applyDesignMetaData(this.$item, false);
		this.loadSelector();
		this.capabilityMaker.apply(this);
		syra_dd.colResizer.push(this);
		this.ensureLayoutMode();
		this.layoutSlot.appendChild(this.domItem);
	},
	initializeList: function() {
		this.$item.$isTopLabelAlignment = true;
		this.RecordClass = LandingPageRecord;
		if (!this.$item.$layout || !this.$item.$layout.$items) {
			this.$item.$layout = this.defineDefaultCard().$layout;
		}
		this.$item.$alternateStyle = false;
		MultiList.prototype.initializeList.call(this);
	},
	fillList: function(dataRecordSet, parentDataRecord, isDelta) {
		MultiList.prototype.fillList.call(this, dataRecordSet, parentDataRecord, isDelta);
		this.applyOrderPreferences();
	},
	isDraggable: function(target, event) {
		return this.isReorderRecordDraggable(target, event);
	},
	dispose: function() {
		syra_site.dom.empty(this._core);
		MultiList.prototype.dispose.call(this);
	}
});
