// syracuse-ui/lib/page/landing/master/landingPageBar.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var BarSplitter = require("syracuse-ui/lib/page/bar/barSplitter").BarSplitter;

function LandingPageBar() {}

exports.LandingPageBar = helpers.defineClass(LandingPageBar, BarSplitter, {
	load: function(page) {
		this.preferenceKey = "ldpbar";
		this.options = {
			resizeDirection: "left",
			$viewMode: "dock",
			$skins: {
				bar: "s-ldpm-bar",
				openButton: "s-ldpm-bar-opener",
				resizePicker: "s-ldpm-bar-resizer"
			}
		};
		this.barSlot = document.createElement("div");
		this.barSlot.className = "s-ldpm-bar-slot";
		this.barBody = document.createElement("div");
		this.barBody.className = "s-ldpm-bar-body";
		if (page.$facet != "$landing_edit") {
			page.domItem.insertBefore(this.barSlot, page.domItem.firstChild);
		}
		BarSplitter.prototype.load.call(this, page);
		this.page.loadNewItem(this.barBody, {
			$isTitleHidden: true,
			$bind: "rolePages"
		});
		if (syra_site.userProfile.hasDesignRight()) {
			var userField = this.page.loadNewItem(this.barBody, {
				$title: syra_local.ldpUserPages,
				$bind: "userPages"
			});
			userField.domItem.className += " s-ldpm-user-pages";
			userField.domTitle.className += " s-ldpm-user-pages";
		}
	}
});
