// syracuse-ui/lib/page/landing/master/landingPageBar.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var BarSplitter = require("syracuse-ui/lib/page/bar/barSplitter").BarSplitter;

function LandingPageBar() {}

exports.LandingPageBar = helpers.defineClass(LandingPageBar, BarSplitter, {
	load: function(page) {
		this.options = {
			resizeDirection: {
				left: true
			},
			$skins: {
				bar: "s-ldp-modulebar",
				openerPicker: "s-ldp-modulebar-opener",
				resizeBar: "s-ldp-modulebar-resizer"
			}
		};
		this.barSlot = document.createElement("div");
		this.barSlot.className = "s-ldp-modulebar-slot";
		this.barBody = document.createElement("div");
		this.barBody.className = "s-ldp-modulebar-body";
		this.barBody.setAttribute("data-s-inout", page.id);
		this.isSlotVisible = true;
		page.domItem.insertBefore(this.barSlot, page.domItem.firstChild);
		BarSplitter.prototype.load.call(this, page);
	},
	onBarBodyAdded: function() {
		this.page.loadNewItem(this.barBody, {
			$isTitleHidden: true,
			$bind: "rolePages"
		});
		this.page.loadNewItem(this.barBody, {
			$isTitleHidden: true,
			$bind: "userPages"
		});
	},
	getArticle: function() {
		return this.page.getArticle();
	}
});
