// syracuse-ui/lib/authoring/tree/contentTree.js
"use strict";
var helpers = require('syracuse-core').helpers;
var ItemTree = require("syracuse-ui/lib/authoring/tree/itemTree").ItemTree;

function ContentTree() {}

exports.ContentTree = helpers.defineClass(ContentTree, ItemTree, {
	load: function(designer) {
		ItemTree.prototype.load.call(this, designer, syra_local.aw_viewTypeContent);
	},
	fill: function() {
		this._clear();
		this.nodes = [];
		this.sectionsNodes = [];
		this.fieldsNodes = [];
		this._addNode(this.designer.designedArticle, {
			node: this.tabItem.body,
			isRoot: true
		});
	},
	_addNode: function(item, parentNode) {
		var items;
		if (!item.isSpaceBox) {
			if (item.isLayout) {
				items = item.items;
			} else {
				//item.parent = parent;
				this._setParentNode(parentNode);

				var itemNode = this._createNodeItem(item.$designLevel);
				parentNode.children.push(itemNode);
				parentNode.childRoot.appendChild(itemNode.node);
				itemNode.item = item;
				this.nodes.push(itemNode);
				if (this.designer.designedItem == item) {
					itemNode.title.className += " s-aw-tree-designed-" + item.$designLevel;
				}
				if (!item.getTitle) {
					//temp udpate waiting extensibility
					item.getTitle = function() {
						return "";
					};
				}
				var title = item.getTitle && item.getTitle(true);
				if (item.$field) {
					this._addFieldType(itemNode, item.$field);
					this.fieldsNodes.push(itemNode);
				} else {
					this.sectionsNodes.push(itemNode);
				}
				this._setTitleNode(itemNode, item.$item, item.$field, title);
			}
			if (item.layoutContent) {
				items = item.layoutContent.items;
			}
			if (items) {
				for (var ii = 0, jj = items.length; ii < jj; ii++) {
					this._addNode(items[ii], itemNode || parentNode);
				}
			}
		}
	}
});
