// syracuse-ui/lib/authoring/tree/itemTree.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

var _iconFields = {
	"application/x-array": "a",
	"application/x-boolean": "b",
	"application/x-choice": "c",
	"application/x-date": "d",
	"application/x-time": "e",
	"application/x-datetime": "e",
	"application/x-decimal": "f",
	"application/x-integer": "f",
	"application/x-real": "f",
	"application/x-graph": "g",
	"application/x-cube": "g",
	"application/x-icon": "h",
	"application/x-binary": "h",
	"application/x-document": "h",
	"image": "h",
	"application/x-quantity": "j",
	"application/x-reference": "j",
	"text/rtf": "k",
	"text/plain": "k",
	"text/html": "k",
	"application/x-password": "l",
	"application/x-string": "l"

	/*"application/x-calendar": require('syracuse-ui/lib/field/schedule/calendar').Calendar,
     "application/x-gantt": require('syracuse-ui/lib/field/schedule/gantt').Gantt,
     "application/x-process": require('syracuse-ui/lib/field/process/visualProcess').VisualProcess,
     */
};

function ItemTree() {}

exports.ItemTree = helpers.defineClass(ItemTree, null, {
	load: function(toolbar) {
		if (this.trees) {
			this.dispose();
		}
		this.toolbar = toolbar;
		this.trees = {};
		if (toolbar.newInsertTab) {
			this._fillInsertTree();
		}
		this.trees.articleTree = {
			nodes: []
		};
		toolbar.newArticleTab.body.style.display = "none";
		document.site.emptyDom(toolbar.newArticleTab.body);
		toolbar.newArticleTab.body.syraAwTree = "articleTree";
		this._addNode(this.toolbar.awPalette.awArticle, {
			node: toolbar.newArticleTab.body,
			isRoot: true
		});
		toolbar.newArticleTab.body.style.display = "";
	},
	findField: function($bind) {
		for (var ii = 0, jj = this._nodes.length; ii < jj; ii++) {
			if (this._nodes[ii].$item.$bind == target.$bind) {
				return this._nodes[ii];
			}
		}
	},
	findNode: function(event, id) {
		var nodes;
		if (id === undefined) {
			var target = event.target;
			while (target && !target.syraNodeId) {
				target = target.parentNode;
			}
			if (target && target.syraNodeId) {
				id = target.syraNodeId;
				nodes = this.trees[this.findSlot(target).syraAwTree].nodes;
			}

		} else {
			nodes = this.trees.articleTree.nodes;
		}
		if (id !== undefined) {
			for (var ii = 0, jj = nodes.length; ii < jj; ii++) {
				if (nodes[ii].node.syraNodeId == id) {
					return nodes[ii];
				}
			}
		}
		return null;
	},
	findSlot: function(target) {
		while (target != null && !target.syraAwTree) {
			target = target.parentNode;
		}
		return target;
	},
	findTree: function(event) {
		var target = this.findSlot(event.target);
		return target.syraAwTree ? this[target.syraAwTree] : null;
	},
	onNodeEvent: function(event) {
		var authorPage = document.site.authorPage;
		delete document.site.requestedDDAuthoringItem;
		event.stopPropagation();
		var node = this.findNode(event);
		if (node) {
			if (event.target.className.indexOf("s-aw-tree-item-opener-children") >= 0) {
				var open = !(node.opener.className.indexOf("s-open") < 0);
				node.children.style.display = open ? "" : "none";
				document.site.toggleClass(node.opener, "s-open", !open);
			} else {
				//event hidden  if card
				switch (event.type) {
					case "mouseenter":
						if (!node.item.$awAddNewItem) {
							authorPage.toggleOverItem(node.item, true);
							this.toolbar.awPalette.targetPage.scrollToItem(node.item);
						}
						break;
					case "mouseleave":
						if (!node.item.$awAddNewItem) {
							authorPage.toggleOverItem(node.item, false);
						}
						break;
					case "click":
						if (!node.item.$awAddNewItem) {
							if (node.item == this.toolbar.awPalette.awArticle) {
								this.toolbar.awPalette.targetPage.scrollToItem(node.item);
							}
							this.selectNode(node.item, true);
							this._isSelecteDisabled = true;
							authorPage.selectItem(node.item, true);
							this._isSelecteDisabled = false;
						}
						break;
				}
			}
		}
	},
	selectNode: function(item, select) {
		if (!this._isSelecteDisabled && item && !item.disposed) {
			var node = this.findNode(null, item.id);
			if (node) {
				if (select) {
					if (this.selectedNode && this.selectedNode != node && this.selectedNode.item) {
						document.site.toggleClass(this.selectedNode.title, "s-aw-tree-designed-" + this.selectedNode.item.$authoringLevel, false);
					}
					this.selectedNode = node;
					document.site.toggleClass(node.title, "s-aw-tree-designed-" + node.item.$authoringLevel, true);
					this.toolbar.awPalette.scrollToItem(this.selectedNode.node, $(this.findSlot(this.selectedNode.node)), true);
				} else {
					if (this.selectedNode == node) {
						this.selectedNode = null;
						document.site.toggleClass(node.title, "s-aw-tree-designed-" + node.item.$authoringLevel, false);
					}
				}
			}
		}
	},
	_addNode: function(item, parentNode) {
		var items;
		if (!item.isSpaceBox) {
			if (item.isLayout) {
				items = item.items;
			} else {
				//item.parent = parent;
				this._setParentNode(parentNode);

				var itemNode = this._addNodeItem(item.id, item.$authoringLevel, parentNode);
				itemNode.item = item;
				this.trees.articleTree.nodes.push(itemNode);

				if (this.toolbar.awPalette.awItem == item) {
					itemNode.title.className += " s-aw-tree-designed-" + item.$authoringLevel;
				}
				var title = item.getTitle();
				if (title == "-" && item.getDefaultTitle) {
					title = item.getDefaultTitle();
				}
				if (item.$field) {
					this._addFieldType(itemNode, item.$field);
				}
				this._setTitleNode(itemNode, item.$item, item.$field, title);
			}
			if (item.layoutContent) {
				items = item.layoutContent.items;
			}
			if (items) {
				for (var ii = 0, jj = items.length; ii < jj; ii++) {
					this._addNode(items[ii], itemNode || parentNode);
				}
			}
		}
	},
	_setParentNode: function(parentNode) {
		if (!parentNode.children) {
			parentNode.children = document.createElement("ul");
			parentNode.children.className = "s-aw-tree-level";
			parentNode.node.appendChild(parentNode.children);
			if (!parentNode.opener) {
				if (parentNode.title) {
					parentNode.opener = document.createElement("a");
					parentNode.opener.className = "s-aw-tree-item-opener";
					parentNode.title.parentNode.insertBefore(parentNode.opener, parentNode.title);
					parentNode.opener.className = "s-aw-tree-item-opener-children";
				}
			}
			if (parentNode.isRoot) {
				parentNode.children.className += " s-aw-tree-root";
			}
		}
	},
	_addNodeItem: function(id, $authoringLevel, parentNode) {
		var itemNode = {};
		itemNode.node = document.createElement("li");
		itemNode.node.syraNodeId = id;
		itemNode.node.className = "s-aw-tree-item s-aw-add-item";
		parentNode.children.appendChild(itemNode.node);

		itemNode.data = document.createElement("div");
		itemNode.data.className = "s-aw-tree-item-data";
		itemNode.node.appendChild(itemNode.data);

		itemNode.title = document.createElement("div");
		itemNode.title.className = "s-aw-tree-item-title s-aw-tree-" + (itemNode.$authoringLevel = $authoringLevel);
		itemNode.titleText = document.createElement("div");
		itemNode.titleText.className = "s-aw-tree-item-text";
		itemNode.title.appendChild(itemNode.titleText);
		itemNode.data.appendChild(itemNode.title);
		return itemNode;
	},
	_setTitleNode: function(itemNode, $item, $field, title) {
		var bindTitle = "",
			desc = title;
		if ($item.$bind) {
			bindTitle = this.toolbar.awPalette.getBindTitle($item.$bind, $field);
			if (title == "") {
				desc = title = bindTitle;
			} else {
				desc = title + " (" + bindTitle + ")";
			}
		}
		itemNode.titleText.textContent = title;
		itemNode.node.title = desc;
	},
	_fillInsertTree: function() {
		this.toolbar.newInsertTab.body.syraAwTree = "insertTree";
		this.toolbar.newInsertTab.body.style.display = "none";
		document.site.emptyDom(this.toolbar.newInsertTab.body);
		this.trees.insertTree = {
			nodes: []
		};
		this._addInsertNode(this.toolbar.awPalette.awArticle.$item, {
			node: this.toolbar.newInsertTab.body,
			isRoot: true
		});
		this._addInsertNewSectionBlockNode("block");
		this._addInsertNewSectionBlockNode("section");
		this.toolbar.newInsertTab.body.style.display = "";
	},
	_addFieldType: function(itemNode, $field) {
		itemNode.type = document.createElement("div");
		itemNode.type.textContent = _iconFields[$field.$type];
		itemNode.type.className = "s-aw-tree-field-type";
		itemNode.titleText.parentNode.insertBefore(itemNode.type, itemNode.titleText);
	},
	_addInsertNewSectionBlockNode: function($authoringLevel) {
		var itemNode = this._addNodeItem("node-" + this.trees.insertTree.nodes.length, $authoringLevel, this.trees.insertTree.nodes[0]);
		var title = $authoringLevel == "section" ? document.site.authorPage.localize.aw_addSection : document.site.authorPage.localize.aw_addBlock;
		itemNode.titleText.textContent = itemNode.node.title = title;
		this.trees.insertTree.nodes.push(itemNode);
		itemNode.item = {
			$title: title,
			$awAddNewItem: true,
			$authoringLevel: itemNode.$authoringLevel
		};
		this.trees.insertTree.nodes[0].children.insertBefore(itemNode.node, this.trees.insertTree.nodes[0].children.firstChild);
	},
	_addInsertNode: function($item, parentNode) {
		var $prototype = this.toolbar.awPalette.awArticle.page.$prototype;
		var $items;
		if ($item.$category != "space") {
			if ($item.$items && !$item.$bind && !$item.$category) {
				$items = $item.$items;
			} else {
				this._setParentNode(parentNode);

				var $authoringLevel = parentNode.isRoot ? "article" : ($item.$bind ? "field" : parentNode.$authoringLevel == "section" ? "block" : "section");
				var itemNode = this._addNodeItem("node-" + this.trees.insertTree.nodes.length, $authoringLevel, parentNode);
				itemNode.$item = $item;
				this.trees.insertTree.nodes.push(itemNode);

				var $field = $item.$bind ? $prototype.$properties[$item.$bind] : null;
				var title;
				if (parentNode.isRoot) {
					title = "Available";
				} else {
					title = $item.$title || ($field ? $field.$title : null) || "";
					if (title && title.indexOf("{") >= 0) {
						if (title[1] == "@") {
							title = title.slice(1, title.length - 1);
							if ($prototype.$localization) {
								title = $prototype.$localization[title];
							}
						}
					}
				}
				if (title == "-") {
					title = document.site.authorPage.localize["aw_" + itemNode.$authoringLevel];
				}
				this._setTitleNode(itemNode, $item, $field, title);
				if ($field) {
					this._addFieldType(itemNode, $field);
					var boundField = this.toolbar.awPalette.awArticle.boundFields[$item.$bind];
					if (boundField && boundField.length > 0) {
						itemNode.item = boundField[0];
					} else {
						itemNode.item = {
							$title: itemNode.node.title,
							$awAddNewItem: true,
							$fieldBind: $item.$bind,
							$fieldCategory: "field"
						};
					}
				} else {
					itemNode.item = {
						$title: itemNode.node.title,
						$awAddNewItem: true,
						$authoringLevel: itemNode.$authoringLevel
					};
				}
			}
			if ($item.$layout && !$item.$bind) {
				$items = $item.$layout.$items;
			}
			if ($items) {
				for (var ii = 0, jj = $items.length; ii < jj; ii++) {
					this._addInsertNode($items[ii], itemNode || parentNode);
				}
			}
		}
	},
	_clearTree: function(tree) {
		if (tree && tree.nodes) {
			for (var ii = 0, jj = tree.nodes.length; ii < jj; ii++) {
				var node = tree.nodes[ii];
				if (node) {
					node.item = node.node = node.children = node.data = null;
					node.opener = node.title = node.titleText = null;
				}
			}
		}
	},
	dispose: function() {
		if (this.toolbar) {
			if (this.toolbar.newArticleTab) {
				this.toolbar.newArticleTab.body.style.display = "none";
				document.site.emptyDom(this.toolbar.newArticleTab.body);
			}
			if (this.toolbar.newInsertTab.body) {
				this.toolbar.newInsertTab.body.style.display = "none";
				document.site.emptyDom(this.toolbar.newInsertTab.body);
			}
			if (this.trees) {
				this._clearTree(this.trees.articleTree);
				this._clearTree(this.trees.insertTree);
			}
			this.trees = this.selectedNode = this.toolbar = null;
		}
	}
});
