// @sage/syracuse-react/src/shortcut/keySequence.ts
"use strict"
import * as keyMap from './keyMap';

export class KeySequence {
    private _combinations: string[]
    private _map: any;
    private _isEnabled: boolean;

    constructor() {
        this.reset();
    }
    get esc(): boolean {
        return this._map.esc;
    }
    get escOnly(): boolean {
        return this.esc && this._combinations.length == 1;
    }
    get tab(): boolean {
        return this._map.tab;
    }
    get space(): boolean {
        return this._map.space;
    }
    get map(): any {
        return this._map;
    }
    get isEnabled(): boolean {
        return this._isEnabled;
    }
    get combinations(): string[] {
        return this._combinations;
    }
    onKeyup(event: KeyboardEvent) {
        let keyName = keyMap.codes[event.keyCode];
        delete this._map[keyName];
        let index = this._combinations.indexOf(keyName);
        this._combinations.splice(index, 1);
    }
    onKeydown(event: KeyboardEvent) {
        if (event.keyCode == keyMap.names.esc) {
            this.reset();
            this._map.esc = true;
            this._combinations = ["esc"];
            event.preventDefault();
            event.stopPropagation();
        } else {
            if (this._map.esc) {
                this._isEnabled = true;
            }
        }
        let keyName = keyMap.codes[event.keyCode];
        if (!this._map[keyName]) {
            this._combinations.push(keyName);
            this._map[keyName] = true;
        }
        return keyName;
    }
    log() {
        console.log("keySequence: map=" + Object.keys(this.map).join(",")) +
            " sequences=" + this._combinations.join(",");
    }
    reset() {
        this._isEnabled = false;
        this._map = {};
        this._combinations = [];
    }

}
