// @sage/syracuse-react/src/profile/temp/technicalLinks.tsx
"use strict";
import * as React from 'react';
import * as Sdata from '../../dataStore/sdata';
import { Link } from '../../link/link';
import { PopupLink } from '../../over/popup/popupLink';
import { ScrollList } from '../../common/list/list';
import { route } from '../../route/route';

export class TechnicalLinks extends React.Component<{
	developpementMode: boolean,
	technicalLinks: Sdata.IMenu[]
}, {}>{
	onClick = (link: Sdata.IMenu) => {
		route.push(link.$url);
	}
	onRenderList = () => {
		return <ScrollList className="s_profile_bar_list">
			{this.props.technicalLinks.map((link: Sdata.IMenu, index: number) =>
				<Link
					key={index}
					className="s_profile_bar_list_item"
					title={link.$title}
					description={link.$description}
					value={link}
					onClick={this.onClick} />)}
		</ScrollList>;
	}
	render() {
		if (this.props.developpementMode && this.props.technicalLinks.length) {
			return <PopupLink
				icon="bug"
				title="Technical Links"
				className="s_profile_bar_textlink"
				popup={{
					my: "middle top",
					at: "right bottom",
					render: this.onRenderList
				}} />;
		}
		return null;
	}
}
