// syracuse-tablet/html/js/controls/panels/aside/panelActions.js
"use strict";

var Base = require('syracuse-tablet/html/js/controls/panels/aside/panelAsideBase').PanelAsideBase;
var locale = require('syracuse-tablet/html/js/helpers/locale');
var utils = require('syracuse-tablet/html/js/helpers/utils');

var panelsHtml = require('syracuse-tablet/html/js/controls/panels/panelsHtml');

var _html = {
	content: '\
		<div class="container-fluid">\
			<div class="navbar-header">\
				<span class="navbar-brand" href="#"><span>{{title}}</span></a>\
			</div>\
		</div>\
		{{{itemsBlockHtml}}}',
	itemsBlock: '\
		<ul class="nav navbar-nav">\
			{{{itemsHtml}}}\
		</ul>',
}

var _templates = {};

function _getHtml(name, ctx) {
	var tmpl = _templates[name];
	if (!tmpl) {
		tmpl = _templates[name] = Handlebars.compile(_html[name]);
	}
	return tmpl(ctx);
};

/**
 * 
 */
exports.PanelActions = utils.defineClass(
	function PanelActions(controller, type, article, opts) {
		Base.call(this, controller, type, article, opts);
	}, Base, {

		destroy: function() {
			Base.prototype.destroy.call(this);
		},

		///
		/// START: Rendering
		///

		/**
		 * 
		 */
		buildHtml: function() {
			var self = this;
			Base.prototype.buildHtml.call(this, "right");

			var items = this.actionAdapter.getActionsPanelItems();
			var links = [];
			var actions = [];
			items.forEach(function(item) {
				if (item.slot === "links") {
					links.push(panelsHtml.getItemHtml("actionPanel", item));
				} else {
					actions.push(panelsHtml.getItemHtml("actionPanel", item));
				}
			});

			if (links.length != 0 && actions.length != 0) {
				links.push('<li class="s-m-separator"></li>');
			}
			var itemsHtml = links.join("") + actions.join("");
			var itemsBlockHtml = "";
			if (itemsHtml.length > 0) {
				itemsBlockHtml = _getHtml("itemsBlock", {
					itemsHtml: itemsHtml
				});
			}
			self.$$panelContent.html(_getHtml("content", {
				title: locale.text("actions"),
				itemsBlockHtml: itemsBlockHtml
			}));
		}
	});
