// syracuse-tablet/html/js/controls/panels/aside/panelAside.js
"use strict";

var Base = require('syracuse-tablet/html/js/controls/panels/aside/panelAsideBase').PanelAsideBase;
var utils = require('syracuse-tablet/html/js/helpers/utils');
var panelsHtml = require('syracuse-tablet/html/js/controls/panels/panelsHtml');
var wpHelpers = require('syracuse-tablet/html/js/sdata/wpHelpers');


var _html = {
	content: '\
		<div class="container-fluid">\
		{{#if htmlHeader}}\
			<div class="navbar-header">\
				<span class="s-m-profile-photo"></span>\
				<span class="navbar-brand" href="#"><span>{{username}}</span>\
				<ul class="nav navbar-nav s-m-panel-slot {{slot}}">\
				{{{htmlHeader}}}\
				</ul>\
			</div>\
		{{/if}}\
		</div>\
		{{{itemsBlockHtml}}}',
	itemsBlock: '\
		<ul class="nav navbar-nav s-m-panel-slot {{slot}}">\
		</ul>',
}

var _templates = {};

function _getHtml(name, ctx) {
	var tmpl = _templates[name];
	if (!tmpl) {
		tmpl = _templates[name] = Handlebars.compile(_html[name]);
	}
	return tmpl(ctx);
};

exports.PanelAside = utils.defineClass(
	function PanelAside(controller, type, article, opts) {
		Base.call(this, controller, type, article, opts);
	}, Base, {

		destroy: function() {
			Base.prototype.destroy.call(this);
		},
		buildHtml: function() {
			var self = this;
			Base.prototype.buildHtml.call(this);
			var ctx = this._getContext();
			self.$$panelContent.html(_getHtml("content", ctx))
			wpHelpers.isPagePinned(self.page.getPinPageId())
				.then(function(state) { //force notif on create panel
					self.notifPinStatusChange(state)
				})
		},
		_getContext: function() {
			var self = this;
			var availableSlots = this._getAvailableSlots();
			var ctx = {},
				slots = {};
			var items = self._filterItems(self._getItems());
			items.forEach(function(item) {
				if (item.slot === "header") {
					ctx.htmlHeader = self._getItemHtml(item);
					ctx.slot = item.slot
				} else {
					item.slot = item.slot || "unclassified"
					if ($.inArray(item.slot, availableSlots) === -1) {
						item.slot = "unclassified"
					}
					slots[item.slot] = slots[item.slot] || [];
					slots[item.slot].push(item);
				}
			});
			var $$itemsBlocksHtml = $("<div></div>");
			availableSlots.forEach(function(slotName) {
				var slot = slots[slotName];
				if (slot) {
					var $$itemsBlock = $(_getHtml("itemsBlock", {
						slot: slotName
					}));
					slot.forEach(function(item) {
						$$itemsBlock.append(self._getItemHtml(item))
					})
					$$itemsBlocksHtml.append($$itemsBlock);
				}

			})
			ctx = this._extendContext(ctx);
			ctx.itemsBlockHtml = $$itemsBlocksHtml.html();
			return ctx;
		},
		_getItemHtml: function(item) {
			return panelsHtml.getItemHtml("globalPanel", item)
		},
		_getItems: function() {
			//to overwrite
			return []
		},
		_filterItems: function(items) {
			var self = this;
			return items.filter(function(item) {
				if (!(self.controller.isActionHidden && self.controller.isActionHidden(item.action))) {
					return true;
				}
			})
		},
		_extendContext: function(ctx) {
			//to overwrite if necessary
			return ctx;
		},
		_getAvailableSlots: function() {
			return ["unclassified"]
		}
	});
