// syracuse-tablet/html/js/controls/panels/aside/panelDashboardTemplate.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var Base = require('syracuse-tablet/html/js/controls/panels/aside/panelAsideDialog').PanelAsideDialog;
var fontUtils = require('syracuse-tablet/html/js/ui/fontUtils');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var notifications = require('syracuse-tablet/html/js/helpers/notifications');
var ModalConfirm = require('syracuse-tablet/html/js/ui/modals/modalConfirm').Modal;

var _templateAll = {
	content: ' \
			<article style="height:100%">\
				<section class="panel panel-default list">\
					<div class="s-m-list-wrapper">\
					</div>\
				</section>\
			{{#if createLabel}}\
				<section class="panel panel-default action">\
					<button type="button" class="btn-primary" data-nevent data-naction="createTemplate"> {{createLabel}}</button>\
				</section>\
			{{/if}}\
			</article>\
		',
	list: ' \
		<ul class="list-group s-m-scroll-elmt">\
		{{#each views}} \
		<li class="list-group-item {{selected}}" data-nevent data-naction="clickView" data-params="{{uuid}}">\
			<table>\
				<tr>\
				<td>\
					<h4 class="list-group-item-heading">{{description}}</h4>\
					<p class="list-group-item-text">{{factoryOwner}}</p>\
				</td>\
				<td>\
					{{#if deletable}}\
					<span  class="btn btn-default s-m-delete-template" data-nevent data-naction="deleteTemplate" data-params="{{uuid}}">\
					<i class="{{../deleteicon}}"></i>\
					</span>\
					{{/if}}\
			    </td>\
				</tr>\
			</table>\
		</li>\
		{{/each}}\
		</ul>'
};
var _getHtml = function(name, ctx) {
	var key = name + "compiled";
	var tmpl = _templateAll[key];
	if (!tmpl) {
		tmpl = _templateAll[key] = Handlebars.compile(_templateAll[name]);
	}
	if (!tmpl) alert("template " + name + " not found");
	return tmpl(ctx);
};

exports.PanelDashboardTemplate = utils.defineClass(
	function PanelDashboardTemplate(controller, $type, article, options) {
		options = options || {};
		options.title = options.title || locale.text("welcome.dashboard.choose.template");
		Base.call(this, controller, $type, article, options);
		this._viewList(options.views);
		notifications.subscribe(this, ["sm.dashboard.list.views.change"])
	}, Base, {
		buildHtml: function() {
			Base.prototype.buildHtml.call(this);
			this.$$content = this._buildContent().appendTo(this.$$panelContent);
			this._buildList().appendTo($(".s-m-list-wrapper", this.$$content))
		},
		destroy: function() {

		},
		_setData: function() {

		},
		_getContext: function() {
			return null;
		},
		_buildContent: function() {
			var opt = {
				"ctrlId": this.id
			};
			if (this.options.showCreateTemplate) {
				opt.createLabel = locale.text("dashboard.welcome.create.template")
			}

			return $(_getHtml("content", opt));
		},
		///
		/// END: Rendering
		///
		show: function() {
			if (!this.$$elmt) {
				this.buildHtml("right");
			}
			this._makeVisible();
		},
		hide: function() {
			if (this.changeCurrentView) {
				this.changeCurrentView = false;
				notifications.publish(["sm.dashboard.change.current.view"])
			};
			Base.prototype.hide.call(this);
		},
		_buildList: function() {
			var opts = {
				"deleteicon": fontUtils.getIconByName("$delete"),
				"ctrlId": this.id,
				"views": this.views
			}
			return $(_getHtml("list", opts));
		},
		_viewList: function(views) {
			var self = this;
			self.views = [];
			views.forEach(function(view) {
				self.views.push({
					"description": view.$description,
					"personal": view.$isPersonal,
					"selected": (view.$selected ? "active" : ""),
					"uuid": view.$uuid,
					"factoryOwner": view.$factoryOwner,
					"deletable": self.options.showCreateTemplate && !view.$isPersonal
				});
			});
		},
		notifDashboardListViewsChange: function(opts) {
			this.changeCurrentView = opts.changeCurrentView;
			this._viewList(opts.views);
			this._replaceList();
		},
		_replaceList: function() {
			$(".s-m-list-wrapper").empty();
			this._buildList().appendTo($(".s-m-list-wrapper", this.$$content))
		},
		_actCreateTemplate: function() {
			this.hide();
			notifications.publish(["sm.dashboard.create.template"]);
		},
		_actDeleteTemplate: function(uuid) {
			var modal = new ModalConfirm(locale.text("modal.confirm.delete.template"), null, true);
			return modal.show().then(function(result) {
				if (result === "yes") {
					notifications.publish(["sm.dashboard.delete.template"], uuid);
				}
			});
		},
		_actClickView: function(uuid) {
			this.hide();
			notifications.publish(["sm.dashboard.select.template"], uuid);
		}
	}
)
