// syracuse-ui/lib/navigation/navigatorPage.js
"use strict";
var helpers = require('syracuse-core').helpers;
var DesktopPage = require("syracuse-ui/lib/page/desktopPage").DesktopPage;
var _module = require('./items/module');

function NavigatorPage() {}

exports.NavigatorPage = helpers.defineClass(NavigatorPage, DesktopPage, {
	load: function(initData) {
		this.navMap = {};
		var activePage = syra_site.getActivePage();
		this.isActivePageFusion = activePage && activePage.isFusionPage;

		this.$item = {
			$layout: {
				$items: []
			}
		};

		syra_pageBuilder.initialize(this);
		this.externalAdapter = syra_site.externalAdapter;
		syra_article.beforeDraw(this);
		this.draw(initData);
		syra_article.endDraw(this);

		this.addChildren();
		this.areModulesLoaded = true;
	},
	addChildren: function() {
		var modules = this.dataset && this.dataset.modules;
		this.children = [];
		if (modules) {
			for (var ii = 0, jj = modules.length; ii < jj; ii++) {
				this.addChild(modules[ii]);
			}
		}
	},
	addChild: function(record) {
		var child = new _module.Module(this, record);
		this.children.push(child);
		return child;
	},
	clear: function() {
		syra_dom.empty(this.mainPopup);
		syra_dom.empty(this.modulesSlot);
		this.lastSelected = null;
		if (this.children) {
			for (var ii = 0, jj = this.children.length; ii < jj; ii++) {
				this.disposeChild(this.children[ii]);
			}
		}
	},
	disposeChild: function(child) {
		if (child) {
			if (child.children) {
				for (var ii = 0, jj = child.children.length; ii < jj; ii++) {
					this.disposeChild(child.children[ii]);
				}
			}
			child.carousel && child.carousel.dispose();
			syra_item.unregister(child);
			syra_site.disposeObject(child);
		}
	},
	dispose: function() {
		this.clear();
		this.mode && this.mode.page.dispose(this);
		DesktopPage.prototype.dispose.call(this);
	}
});
