// etna/lib/engine/compilers/nodes.ts

export interface ParseNode {
	tag: string;
}

export interface ParseValueNode extends ParseNode {
	value: any;	
}

export interface ParseMemberNode extends ParseNode {
	target: ParseNode;
	name: string;
}

export interface ParseDictNode extends ParseNode {
	field: string;
	letter: string;
	abbrev: string;
}

export interface ParseTypeNode extends ParseNode {
	field: string;
	letter: string;
	exp: ParseNode;
}

export interface ParseOpNode extends ParseNode {
	op: string;
	args: ParseNode[];
}

export interface ParseFCallNode extends ParseNode {
	name: string;
	args: ParseNode[];
}

export interface ParseFuncNode extends ParseNode {
	from: ParseNode;
	name: ParseNode;
	args: ParseNode[];
}

export interface ParseIndexNode extends ParseNode {
	target: ParseNode;
	args: ParseNode[];
}

export interface ParseArgNode extends ParseNode {
	arg: ParseNode;
}

export interface ParseArgsNode extends ParseNode {
	args: ParseNode[];
}

export interface ParseFieldArgsNode extends ParseNode {
	field: string;
	args: ParseNode[];
}

export interface ParseDateNode extends ParseNode {
	year: number;
	month: number;
	day: number;
}
