// @sage/syracuse-lib/src/collaboration/util.ts
"use strict";

import * as sdataRegistry from '../../src/sdata/sdataRegistry';
import * as dataModel from '../../src/orm/dataModel';
import { helpers } from '@sage/syracuse-core';

var _config = {};

export function setup(config) {
	var collConf = config || {};
	_config.application = collConf.application || "syracuse";
	_config.contract = collConf.contract || "collaboration";
	_config.dataset = collConf.dataset || "syracuse";
};

export function getCollaborationModel() {
	var contract = sdataRegistry.getContract(_config.application, _config.contract, true);
	var dataset = _config.dataset;
	return dataModel.make(contract, dataset);
};

export function getCollaborationOrm(_) {
	var contract = sdataRegistry.getContract(_config.application, _config.contract, true);
	var dataset = _config.dataset;
	//
	return dataModel.getOrm(_, exports.getCollaborationModel(), contract.datasets[dataset]);
};

export function getEndpoints(_, options) {
	var db = exports.getCollaborationOrm(_);
	return db.fetchInstances(_, db.model.getEntity(_, "endPoint"), options);
};

export function getEndpoint(_, options) {
	var opt = {};
	// lookup in config first
	// get db endpoints
	if (options.jsonWhere) opt.jsonWhere = options.jsonWhere;
	else if (options.application && options.contract && options.dataset)
	// case insensitive search
		opt.jsonWhere = {
		application: {
			$regex: "^" + options.application + "$",
			$options: "i"
		},
		contract: {
			$regex: "^" + options.contract + "$",
			$options: "i"
		},
		dataset: {
			$regex: "^" + options.dataset + "$",
			$options: "i"
		}
	};
	var db = exports.getCollaborationOrm(_);
	return db.fetchInstance(_, db.model.getEntity(_, "endPoint"), opt);
};
