// syracuse-ui/lib/mobile/field/collection/cardRecord.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Article = require("syracuse-ui/lib/common/article/article").Article;

function CardRecord(){
}

exports.CardRecord = helpers.defineClass(CardRecord, Article, {
    loadRecord: function(list, record, $recordIndex, $$recordContainer){
        this.$recordUuid = record.$uuid;
        this.$recordIndex = $recordIndex;
        this.$$recordContainer = $$recordContainer;
        this.$facet = list.$recordFacet;
        this.$prototype = list.$prototype.$item;
        document.itemFactory.initializeItem(this, {
            $layout: {
                $items: list.$item.$layout.$items
            }
        }, list);
        this.loadBox(record);
        
    },
    remove: function(){
        this.$$item.remove();
    },
    drawBox: function(){
        var $$container = this.$$item = $("<li/>").appendTo(this.boxParent._$$body).attr("data-s-record", this.$recordUuid).addClass("s-field-cardRecord");
        var $links = this.$prototype.$links || {};
        var $link = $links.$details || $links.$edit;
        if ($link) {
            this.link = document.itemFactory.load(this.$$item, {
                $bind: "$detail",
                $category: "link",
                $isListWrapper: true,
            }, this);
            this.link.setMenu($link);
            $$container = this.link.$$item;
        }
        document.itemFactory.load($$container, this.boxParent.builder.$cardview, this);
    },
    highlightSelection: function(selected){
        this.$$item.toggleClass("s-list-record-selected", selected);
        if (this.$$selector && this.$$selector.is("input")) {
            this.$$selector.attr('checked', selected);
        }
    }
});
