// syracuse-ui/lib/mobile/field/collection/cardListBuilder.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var RawListBuilder = require("syracuse-ui/lib/common/field/collection/rawListBuilder").RawListBuilder;
var localize = require('syracuse-ui/lib/common/controller/localize').list;
var Pager = require("./pager").Pager;
var CardRecord = require('./cardRecord').CardRecord;


function CardsBuilder(){
}

exports.Builder = helpers.defineClass(CardsBuilder, RawListBuilder, {
    initialize: function(){
        this.RecordClass = CardRecord;
		RawListBuilder.prototype.initialize.call(this);
    },
    drawBuilder: function(){
        var self = this;
        var list = self.list;
        list.$skin = list.$item.$skin || "s-grid";
        
        list.appendArticleMenus(list.$$dataValue);
        if (!list.$item.$isPagerHidden) {
            (list._pager = new Pager()).load(list);
        }
        self._body = document.createElement("ul");
        self._body.setAttribute("data-inset", true);
        self._body.setAttribute("data-theme", list.$theme.$item || document.site.$item.$themes.$default.$item);
        self._body.style.display = "none";
        list.$$dataValue[0].appendChild(self._body);
        
        this.$cards = (typeof(list.$item.$cards) === "object") ? list.$item.$cards : {
            $layout: {
                $items: list.$item.$layout.$items
            }
        };
        this.$cards.$category = "block";
        this.$cards.$isBodyTransparent = true;
        
        list.applyDesignMetaData(list.$item, false);
        list.boxParent.getArticle().bind(list, list.$item.$bind);
        self.refreshListView();
        self._body.style.display = "";
        list._$$body = $(self._body);
    },
    refreshListView: function(){
        if (this.list.getPage().$isjqmPage) {
            document.site.refreshJQMPlugin(this.list._$$body, "listview");
        }
    },
    onAfterDataBinding: function(){
        if (this.list._pager) {
            this.list._pager.draw();
        }
        this.refreshListView();
    }
});
