// syracuse-ui/lib/mobile/field/collection/cardRecord.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Article = require("syracuse-ui/lib/common/article/article").Article;

function CardRecord(){
}

exports.CardRecord = helpers.defineClass(CardRecord, Article, {
    initializeRecord: function(options){
        options.$item = {
            $layout: {
                $items: this.list.$item.$layout.$items
            }
        };
    },
    dispose: function(){
        delete this.list
        delete this.builder;
        delete this.row;
        delete this.menus;
        Article.prototype.dispose.call(this);
    },
    applyMetaData: function(metaData){
    },
    remove: function(){
        this.$$item.remove();
    },
    drawBox: function(){
        var self = this;
        var $$container = self.$$item = $(self.boxParent.builder._body.appendChild(document.createElement("li"))).attr("data-s-record", self.$recordUuid).addClass("s-field-cardRecord");
        var $links = self.$prototype.$links || {};
        var $link = $links.$details || $links.$edit;
        if ($link) {
            self.link = document.itemFactory.load(self.$$item, {
                $bind: "$detail",
                $category: "link",
                $isListWrapper: true,
            }, self);
            self.link.setMenu($link);
            $$container = self.link.$$item;
            
            // if lookup
            if (self.boxParent.page.$facet == "$lookup") {
                $$container.bind("click", function(){
                    var selectedRecords = {};
                    selectedRecords[self.$recordUuid] = self;
                    self.boxParent.page.onSelectRecord(selectedRecords);
                    return false;
                });
            }
        }
        document.itemFactory.load($$container, self.boxParent.builder.$cards, self);
    },
    highlightSelection: function(selected){
        this.$$item.toggleClass("s-list-record-selected", selected);
        if (this.$$selector && this.$$selector.is("input")) {
            this.$$selector.attr('checked', selected);
        }
    }
});
