// syracuse-ui/lib/mobile/field/collection/referenceRecord.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Article = require("syracuse-ui/lib/common/article/article").Article;

function ReferenceRecord(){
}

exports.ReferenceRecord = helpers.defineClass(ReferenceRecord, Article, {
    loadRecord: function(list, record, $recordIndex){
        this.$recordUuid = record.$uuid;
        this.$recordIndex = $recordIndex;
        this.$facet = list.$recordFacet;
        this.$isEditMode = list.$isEditMode;
        this.$prototype = {
            $: {
                refField: list.$prototype.$item
            }
        };
        document.itemFactory.initializeItem(this, {}, list);
        this.loadBox({
            $uuid: record.$uuid,
            refField: record
        });
    },
    remove: function(){
        this.$$item.remove();
    },
    drawBox: function(){
        var self = this;
        var list = self.boxParent;
        self.$$item = $("<div/>").addClass(list.$skin + "-record").attr("data-s-record", self.$recordUuid).addClass(list.$recordCss);
        var $$data = $("<div/>").addClass(list.$skin + "-record-data").appendTo(self.$$item);
        if (this.$isEditMode) {
            // grouped buttons style attribute
            //$$data.attr("data-role", "controlgroup").attr("data-type", "horizontal");
            
            self.$$del = $("<a/>").addClass(list.$skin + "-record-del").appendTo(self.$$item).bind("click", function(){
                document.controller.notifyChange(list.getArticleParent(), list.$item.$bind, [{
                    $uuid: self.$recordUuid,
                    $index: self.$recordIndex,
                    $isDeleted: true
                }]);
                return false;
            });
            //self.$$del.attr("data-theme", document.site.$item.$themes.$default.$button);
            //self.$$del.button(); // necessary in case of refresh
        }
        document.itemFactory.load($$data, {
            $isTitleRowHidden: true,
            $bind: "refField",
            $isEditMode: false,
            $inplace: true
        }, this);
        
        self.$$item.appendTo(list._$$body);
    },
    dispose: function(){
        if (this.$$del) {
            this.$$del.unbind();
        }
        Article.prototype.dispose.call(this);
    }
    
});
