// syracuse-tablet/html/js/controls/chart/chartUserPrefs.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var sessStorage = require('syracuse-tablet/html/js/storage/sessionStorage');

var _Klass = utils.defineClass(function ChartDetailPrefs(id) {
	this._id = id;
}, null, {
	destroy: function() {},
	_toolbarWrite: function(type, info) {
		sessStorage.setItem(this._getStorageId(type), info ? JSON.stringify(info) : null);
	},
	_toolbarRead: function(type) {
		var info = sessStorage.getItem(this._getStorageId(type));
		if (!info) return null;
		try {
			return JSON.parse(info);
		} catch (e) {};
		return null;
	},
	_getStorageId: function(type) {
		return this._id + type;
	},
	getPrefs: function() {
		return this._toolbarRead("toolbar", this) || {};
	},
	savePrefs: function(opts) {
		var prefs = null;
		if (opts !== null) {
			var prefs = this.getPrefs();
			for (var p in opts) {
				prefs[p] = opts[p];
			}
		}
		this._toolbarWrite("toolbar", prefs);
	},
	reset: function(opts) {
		this.savePrefs(null);
	}
});

exports.createChartDetailPrefs = function(id) {
	return new _Klass(id);
};
