// syracuse-tablet/html/js/controls/chart/ctrlChartDetailToolbar.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var fontUtils = require('syracuse-tablet/html/js/ui/fontUtils');
var prototype = require('syracuse-tablet/html/js/helpers/prototype');
var ctrlBase = require('syracuse-tablet/html/js/controls/ctrlBase');
var uiSettings = require('syracuse-tablet/html/js/helpers/uiSettings');

var Base = ctrlBase.Klass;

var _actions = [{
	params: "action1",
	icon: "glyphicon glyphicon-align-left"
}, {
	params: "action2",
	icon: "glyphicon glyphicon-align-left"
}, {
	params: "action3",
	icon: "glyphicon glyphicon-align-left"
}];

var _templates = {
	toolbar: '\
		<div class="btn-toolbar" role="toolbar">\
			<div class="btn-group">\
				{{#each actions}}\
					<button type="button" class="btn btn-default" data-action="toolbar" data-control-id="{{../ctrlId}}" data-params="{{params}}"><span class="{{icon}}"></span></button>\
				{{/each}}\
		    </div>\
	    </div>'

};
var _getHtml = function(name, ctx) {
	var key = name + "compiled";
	var tmpl = _templates[key];
	if (!tmpl) {
		tmpl = _templates[key] = Handlebars.compile(_templates[name]);
	}
	if (!tmpl) alert("template " + name + " not found");
	return tmpl(ctx);
};


/**
 * chartDetail Tool bar
 */
var _Klass = utils.defineClass(

	function CtrlChartDetailToolbar($type, controller, options) {
		Base.call(this, controller, {
			$bind: "chartDetailToolbar"
		}, prototype.create({
			$type: $type
		}), options);

	}, Base, {
		buildHtml: function($$parent, controllerDao, buildOptions) {
			this._height = uiSettings.getProp("chartDetailToolbar.height");
			Base.prototype.buildHtml.call(this, $$parent, controllerDao, buildOptions, [this.typeName]);
			this.$$elmt.html(_getHtml('toolbar', {
				ctrlId: this.id,
				actions: _actions
			}));
			this.$$elmt.css("height", this._height);
		},
		getHeight: function() {
			return this.$$elmt.is(":visible") ? this._height : 0;
		},
		_actToolbar: function(params) {
			var chart = this.controller.getChart();
			var array = this.controller.getArray();
			console.log(params, "chartCtrl", chart ? "found" : "not found", "arrayCtrl", array ? "found" : "not found");
		}
	});

exports.Klass = _Klass;
