// syracuse-ui/lib/site/authoring/authorPage.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var locale = require('syracuse-core/lib/locale');
var RawPage = require("syracuse-ui/lib/article/rawPage").RawPage;
var AreaScroller = require('./areaScroller').AreaScroller;
var TreeNode = require('./treeNode').TreeNode;
var layoutProvider = require('syracuse-ui/lib/article/layouts/layoutProvider');

var palettes = {
    article: require("./palette/pagePalette").PagePalette,
    section: require("./palette/sectionBlockPalette").SectionBlockPalette,
    field: require("./palette/fieldPalette").FieldPalette
};
palettes.block = palettes.section;

function AuthorPage(){
}

exports.AuthorPage = helpers.defineClass(AuthorPage, RawPage, {
    notifyUpdate: function(){
        this.isUpdated = true;
    },
    open: function(){
        this.localize = locale.resources(module)();
        this.targetPage = document.site.mainPage;
        
        this.$sourceItem = helpers.object.clone(this.targetPage.$item, true);
        this.$defaultItem = helpers.object.clone(document.site.ensureDefaultArticle(this.targetPage.$prototype.$article, this.targetPage.$prototype), true);
        this.isUpdated = false;
        document.controller.loadWorkingCopy({
            menu: {
                $url: this.targetPage.$authorUrl
            },
            article: this,
            callback: function(){
                document.site.resize();
            }
        });
    },
    
    onMenuClick: function(menuItem){
        var self = this;
        switch (menuItem.$item.$bind) {
            case "$autoLayout":
                var $newLayout = layoutProvider.buildAutoLayout(this.targetPage.$prototype, helpers.object.clone(this.$sourceItem, true));
                self.targetPage.reloadLayout($newLayout);
                break;
            case "$defaultLayout":
                var dataset = self.dataset;
                self.targetPage.reloadLayout(helpers.object.clone(self.$sourceItem, true));
                break;
            case "$resetLayout":
                self.targetPage.reloadLayout(helpers.object.clone(self.$defaultItem, true));
                break;
            case "$close":
                if (self.isUpdated) {
                    document.site.showMessage({
                        $title: self.localize.aw_updateMessageTitle,
                        $message: self.localize.aw_cancelMessageText,
                        $type: "question",
                        $buttons: "yesnocancel",
                        callback: function(response){
                            if (response.$id == "yes") {
                                self.menuItems.$save[0].click();
                            }
                            else {
                                if (response.$id == "no") {
                                    document.site.toggleAuthoring(false);
                                    document.controller.reloadMainPage();
                                }
                            }
                        }
                    });
                }
                else {
                    document.site.toggleAuthoring(false);
                }
                return false;
                break;
        }
        return true;
        
    },
    drawBox: function(){
        var self = this;
        self.$$pathSteps = {};
        self.$$container.empty().show();
        
        this.nodes = {};
        this.scroller = new AreaScroller(document.site.$$body);
        
        self.$$item = self.$$container;
        var $$header = $("<div id='s-aw-header'/>").appendTo(self.$$container);
        self.$$title = $("<a id='s-aw-header-title' href='#'/>").text(self.localize.aw_title).appendTo($$header);
        self.$$title.bind("click", function(){
            self.$isPaletteExpanded = !self.$isPaletteExpanded;
            self.$$title.toggleClass("s-open", self.$isPaletteExpanded);
            if (self._$$panel) {
                self._ensurePanelVisibility();
                document.site.resize();
            }
            return false;
        });
        
        self.$$switchModeSlot = $("<div id='s-aw-header-switch-mode'/>").appendTo($$header);
        var $$tools = $("<div id='s-aw-header-tools'/>").appendTo($$header);
        self.$$layoutSlot = $("<div id='s-aw-header-layouts'/>").appendTo($$tools);
        self.$$panelActions = $("<div id='s-aw-header-panel-actions'/>").appendTo($$tools);
        var $$body = $("<div id='s-aw-header-body'/>").appendTo($$header);
        var $$actions = $("<div class='s-aw-header-actions'/>");
        /* self.loadNewItem($$actions, {
         $bind: "$autoLayout",
         $category: "link",
         $skin: "s-aw-header-links-link"
         
         });
         self.loadNewItem($$actions, {
         $bind: "$defaultLayout",
         $title: "default",
         $category: "link",
         $skin: "s-aw-header-links-link"
         });
         self.loadNewItem($$actions, {
         $bind: "$resetLayout",
         $category: "link",
         $skin: "s-aw-header-links-link"
         });*/
        self.loadNewItem($$actions, {
            $bind: "$save",
            $category: "link",
            $skin: "s-aw-save"
        });
        self.headerMenus = self.loadNewItem($$actions, {
            $category: "links",
            $skin: "s-aw-header-links",
            $excludeBind: ["$details", "$query", "$print"]
        });
        $$actions.appendTo($$body);
        var close = self.loadNewItem($$body, {
            $bind: "$close",
            $category: "link",
            $noText: true,
            $skin: "s-aw-close"
        });
        close.setMenu({
            $title: self.localize.aw_close
        }, null);
        self.applyChange({
            $actions: {
                $previousLayout: {
                    $title: "prev"
                },
                $nextLayout: {
                    $title: "next"
                },
                $autoLayout: {
                    $title: "auto"
                },
                $defaultLayout: {
                    $title: "default"
                },
                $resetLayout: {
                    $title: "reset"
                }
            }
        });
        self._appendPopupHost();
        
        self.$$authorBody = $("<div id='s-aw-body'/>").appendTo(self.$$container);
        self.appendBreadcrumb(self.$$authorBody);
        self._$$panel = $("<div class='s-aw-panel'/>").appendTo(self.$$authorBody);
        self._ensurePanelVisibility();
        document.site.resize();
        
        (this.rootNode = new TreeNode(self.targetPage, null, self)).load().selectItem();
    },
    _ensurePanelVisibility: function(){
        this._$$panel[0].style.display = this.$isPaletteExpanded ? "" : "none";
    },
    _appendPopupHost: function(){
        var self = this;
        self._$$popup = $("<div class='s-aw-tree-host'/>").delegate("a[data-s-node]", "click", function(){
            self._findNode(this).expand();
            setTimeout(function(){
                document.site.resize()
            }, 10);
            return false;
        }).delegate("input.s-aw-tree-item-check", "click", function(){
            var node = self._findNode(this);
            var state = {
                $isHidden: !$(this).is(":checked")
            };
            node.applyHiddenState(state);
            node.item.applyDesignMetaData(state);
        }).delegate("div.s-aw-tree-item-title", "mouseenter mouseleave click", function(event){
            var node = self._findNode(this);
            switch (event.type) {
                case "mouseenter":
                    node.toggleOverItem(true);
                    if (self._showTargetTimer) {
                        window.clearTimeout(self._showTargetTimer);
                    }
                    self._showTargetTimer = window.setTimeout(function(){
                        delete self._showTargetTimer;
                        self.showTargetItem(node);
                    }, 100)
                    break;
                case "mouseleave":
                    node.toggleOverItem(false);
                    break;
                case "click":
                    self.selectItem(node);
                    break;
            }
            return false;
        });
    },
    selectPathStep: function(node){
        var self = this;
        setTimeout(function(){
            self.togglePopup(false);
            if (self.designedNode) {
                self.removeItem(self.palette);
                if (self.$$pathSteps[self.designedNode.id]) {
                    self.$$pathSteps[self.designedNode.id].toggleClass("s-open", false);
                }
                self.designedNode.highlightDesignedItem(false);
            }
            self.designedNode = node;
            self.showTargetItem(node);
            self.$$pathSteps[node.id].toggleClass("s-open", true);
            self.palette = new palettes[self.designedNode.$authoringLevel]();
            self.palette.localize = self.localize;
            self.palette.authorPage = self;
            self.palette._initializePage();
            self.palette.externalAdapter = document.site.externalAdapter;
            var saveDisplay = self._$$panel[0].style.display;
            (self.palette.$$container = self._$$panel.empty())[0].style.display = "none ";
            self.palette.$layoutOptions = document.site.$item.$layoutOptions.authoring;
            self.palette.loadBox();
            self.palette.$$container[0].style.display = saveDisplay;
            node.highlightDesignedItem(true);
            document.site.resize();
        }, 10);
    },
    appendBreadcrumb: function($$body){
        var self = this;
        self._$$path = $("<nav class='s-aw-path'/>").delegate("a.s-aw-path-link", "click", function(){
            self.selectPathStep(self._findNode(this));
            return false;
        }).delegate("a.s-aw-path-opener", "click", function(event){
            setTimeout(function(){
                self.togglePopup(event.target);
            }, 10);
            return false;
        }).appendTo($$body);
    },
    _drawStep: function(node){
        var dom = document.createElement("a");
        dom.className = "s-aw-path-link s-aw-path-" + node.$authoringLevel;
        dom.setAttribute("href", "#");
        dom.setAttribute("title", this.localize["aw_" + (node.item.page == node.item ? "page" : node.$authoringLevel) + "Step"]);
        node.addNodeId(dom);
        var title = node.item.getTitle();
        if (title.length > 15) {
            dom.setAttribute("title", title);
            title = title.slice(0, 15) + "...";
        }
        else {
            if (title == "") {
                title = node.item.getDefaultTitle();
            }
        }
        var $$step = $(dom);
        $$step.text(title);
        dom = document.createElement("a");
        dom.className = "s-aw-path-opener" + ((node.parentNode && node.parentNode.children.length > 1) ? " s-open" : "");
        node.addNodeId(dom);
        return (this.$$pathSteps[node.id] = $$step.append(dom).toggle(true));
    },
    selectItem: function(selectNode, doSelect){
        if (doSelect !== false) {
            if (doSelect || (selectNode != this.designedNode)) {
                this.$$pathSteps = {};
                this._$$path.empty();
                var curNode = selectNode;
                while (curNode.parentNode) {
                    if (curNode.parentNode.item) { //test root
                        this._drawStep(curNode.parentNode).prependTo(this._$$path);
                    }
                    curNode = curNode.parentNode;
                }
                this._drawStep(selectNode).appendTo(this._$$path);
                curNode = selectNode
                while (curNode.children.length > 0) {
                    this._drawStep(curNode = curNode.children[0]).appendTo(this._$$path);
                }
                this.selectPathStep(selectNode);
            }
        }
    },
    applyChange: function(newData){
        var self = this;
        var $save;
        if (newData && newData.$actions && ($save = newData.$actions.$save)) {
            if ($save.$isDisabled && $save.$isRequested === false) {
                var isFailed;
                if ($save.$diagnoses) {
                    $save.$diagnoses.some(function($diagnoses){
                        return (isFailed = ($diagnoses.severity !== "info"));
                    });
                }
                if (!isFailed) {
                    delete newData.$actions.$save;
                    document.site.showMessage({
                        $title: self.localize.aw_updateMessageTitle,
                        $message: self.localize.aw_saveMessageText,
                        $type: "question",
                        $buttons: "yesno",
                        callback: function(response){
                            if (response.$id != "no") {
                                document.site.toggleAuthoring(false);
                                document.controller.reloadMainPage();
                            }
                            self.isUpdated = false;
                        }
                    });
                }
            }
        }
        
        RawPage.prototype.applyChange.call(this, newData);
    },
    togglePopup: function(target){
        var self = this;
        if (target !== false) {
            var node = self._findNode(target);
            if (self._popupNode && self._popupNode != node) {
                self._popup.close();
                delete self._popup;
            }
            if (!self._popup) {
                if (node != self.designedNode) {
                    self.selectItem(node);
                }
                self._popupNode = node;
                var dom = document.createElement("ul");
                dom.className = "s-aw-tree-root";
                node.parentNode.$$children = $(dom).appendTo(self._$$popup.hide())
                node.parentNode.drawChildNodes(true);
                var $$step = self.$$pathSteps[node.id];
                self._popup = document.site.openDialog({
                    $$dialog: self._$$popup,
                    content: node,
                    $dialogMode: "popup",
                    autocCloseBoundary: "#" + $$step.attr("id"),
                    position: {
                        my: "left top",
                        at: "left bottom",
                        of: $$step
                    },
                    onClose: function(){
                        delete self._popup;
                        delete self._popupNode;
                        self._$$popup.empty();
                        node.parentNode.drawChildNodes(false);
                        delete node.parentNode.$$children;
                        self.showOverPathNodes();
                    }
                });
                node.toggleOverItem(true);
                self.showTargetItem(node);
                return;
            }
        }
        if (self._popup) {
            self._popup.close();
        }
    },
    notifyActionChange: function(target, value, notifyServer){
        this.ensureSendBag().content = {
            $article: this.targetPage.$item
        };
        /*if (!this.dataset.application) {
         var httpQuery = document.controller.parseUrl(this.targetPage.$prototype.$representationUrl);
         data.application = httpQuery.$urlParts.endpointParts.application;
         data.contract = httpQuery.$urlParts.endpointParts.contract;
         data.representation = httpQuery.$urlParts.representationRoot;
         data.facet = httpQuery.$urlParts.$facet;
         }*/
        RawPage.prototype.notifyActionChange.call(this, target, value, notifyServer);
    },
    notifyDataChange: function(field, value){
        this.ensureSendBag().content = {
            $article: this.targetPage.$item
        };
        
        RawPage.prototype.notifyDataChange.call(this, field, value);
    },
    _findNode: function(target){
        return this.nodes[target.getAttribute("data-s-node")];
    },
    showTargetItem: function(itemNode){
        var $$target;
        //ensure tabParent is opened
        var node = itemNode;
        while (node && node.item) {
            if (node.item.$isTabLayout && !node.item.$item.$opened) {
                node.item.openBox(true);
            }
            node = node.parentNode
        }
        var item = itemNode.item;
        if (item.$isTabLayout && !item.$item.$opened) {
            item.openBox(true);
            
        }
        var $$target = item[item.$isTabLayout ? "$$header" : "$$item"];
        if ($$target) {
            if (!item.$isTabLayout) {
                this.scroller.scrollToTarget($$target, true);
            }
        }
    },
    showOverPathNodes: function(overNode, show){
        if (!overNode) {
            overNode = this.overNode;
            show = false;
            delete this.overNode;
        }
        else {
            if (this.overNode != overNode) {
                this.showOverPathNodes();
            }
            this.overNode = overNode;
        }
        while (overNode) {
            overNode.toggleItemCss("over", show);
            overNode = overNode.parentNode;
        }
    },
    
    moveContent: function(options){
        var count = options.sourceNode.children.length - 1;
        var layout = options.layout || (options.action ? options.targetItem.layoutParent : options.targetItem.layoutContent);
        var firstVisibleNode;
        while (count >= 0) {
            var node = options.sourceNode.children[count--];
            if (node.item != options.targetItem) {
                if (node.$authoringLevel == (options.levelMoved || "field")) {
                    if (!firstVisibleNode && !node.$isHidden) {
                        firstVisibleNode = node;
                    }
                    layout.appendNewItem({
                        newItem: node.item,
                        targetItem: options.targetItem,
                        action: options.action,
                        doSelectItem: false
                    });
                }
                else {
                    if (item.layoutContent) {
                        this.moveContent({
                            sourceNode: node,
                            targetItem: options.targetItem,
                            action: options.action,
                            levelMoved: options.levelMoved,
                            layout: options.layout
                        });
                    }
                }
            }
        }
        if (firstVisibleNode && options.doSelectItem !== false) {
            firstVisibleNode.selectItem(true);
        }
    },
    addNewBox: function(options){
        var newBox = options.parentBox.page.createNewItem({
            $category: options.$category,
            $title: this.localize["aw_new" + options.$category.slice(0, 1).toUpperCase() + options.$category.slice(1)],
            $layout: {
                $items: []
            }
        }, options.parentBox);
        (options.layout || options.parentBox.layoutContent).appendNewItem({
            newItem: newBox,
            load: true,
            targetItem: options.targetItem,
            action: options.action
        });
        if (options.openBox && !newBox.layoutContent) {
            newBox.openBox(true);
        }
        return newBox;
    },
    dispose: function(){
        this.showOverPathNodes();
        delete this.targetPage;
        delete this.designedNode;
        if (this.$$title) {
            this.$$title.unbind();
        }
        if (this._$$path) {
            this._$$path.undelegate();
        }
        if (this._$$popup) {
            this._$$popup.undelegate().remove();
        }
        delete this.page;
        RawPage.prototype.dispose.call(this);
    }
});
