// syracuse-collaboration/lib/representations/msoffice/documentExcel.js
"use strict";
exports.representation = {
	resources: function() {
		var locale = require("syracuse-core/lib/locale");
		var globals = require('streamline/lib/globals');
		return locale.resources(module, globals.context && globals.context.sessionLocale)();
	},
	$entityName: "document",
	$facets: {
		$details: {
			$prototype: {
				$links: {
					$query: {
						$isHidden: true
					},
					$delete: {
						$isHidden: true
					}
				},
				$actions: {
					$saveDocument: {
						$title: "{@SaveActionTitle}"
					}
				}
			},
			$layout: {
				$items: [{
					$layoutType: "row",
					$items: [{
						$items: [{
							$bind: "description"
						}, {
							$bind: "volume"
						}, {
							$bind: "documentDate"
						}]
					}, {
						$items: [{
							$bind: "owner"
						}, {
							$bind: "isReadOnly"
						}]
					}]
				}, {
					$layoutType: "stack",
					$items: [{
						$bind: "teams"
					}, {
						$bind: "tags"
					}]
				}]
			}
		},
		$edit: {
			$prototype: {
				$links: {
					$query: {
						$isHidden: true
					}
				},
				$actions: {
					$save: {
						$links: {
							$query: {
								$isHidden: true
							},
							$create: {
								$isHidden: true
							}
						}
					}
				}
			},
			$layout: {
				$items: [{
					$layoutType: "row",
					$items: [{
						$items: [{
							$bind: "description"
						}, {
							$bind: "volume"
						}]
					}, {
						$items: [{
							$bind: "owner"
						}, {
							$bind: "isReadOnly"
						}]
					}]
				}, {
					$layoutType: "stack",
					$items: [{
						$bind: "teams"
					}, {
						$bind: "tags"
					}]
				}]
			}
		},
		$query: {
			$prototype: {
				$links: {
					$create: {
						$title: "{@PublishActionTitle}"
					},
					$print: {
						$isHidden: true
					},
					$excel: {
						$isHidden: true
					}
				}
			},
			$layout: {
				$items: [{
					$category: "section",
					$layout: {
						$items: [{
							$bind: "$resources",
							//                            $format: "excelDocumentCardList",
							$format: "cards",
							$layout: {
								$items: [{
									"$category": "section",
									"$expression": "{@ResourceExpression}"
								}]
							}
						}]
					}
				}]
			}
		}
	}
};
