// syracuse-collaboration/lib/representations/msoffice/msoExcelDocument.js
"use strict";
exports.representation = {
	resources: function() {
		var locale = require("syracuse-core/lib/locale");
		var globals = require('streamline/lib/globals');
		return locale.resources(module, globals.context && globals.context.sessionLocale)();
	},
	$entityName: "document",
	$facets: {
		$query: {
			$prototype: {
				$links: {
					$query: {
						$isHidden: true
					},
					$create: {
						$isHidden: true
					},
					$delete: {
						$isHidden: true
					},
					$excel: {
						$isHidden: true
					},

					$print: {
						$isHidden: true
					},
					$wordmailmerge: {
						$isHidden: true
					},
					$wordreport: {
						$isHidden: true
					},
					saveNewDocument: {
						$title: "Save as document",
						$type: "application/json;vnd.sage=syracuse",
						$url: "/sdata/syracuse/collaboration/syracuse/documents/$template/$workingCopies?representation=msoExcelDocument.$edit",
						$method: "POST"
					},
					saveReportTemplate: {
						$title: "Save as report template",
						$type: "application/json;vnd.sage=syracuse",
						$url: "/sdata/syracuse/collaboration/syracuse/msoExcelTemplateDocuments/$template/$workingCopies?representation=msoExcelReportTemplate.$edit",
						$method: "POST"
					}
				},
			},
			$layout: {
				$items: []
			}
		},
		$edit: {
			$prototype: {
				$links: {
					$query: {
						$isHidden: true
					},
					$print: {
						$isHidden: true
					},
					$wordreport: {
						$isHidden: true
					}
				},
				$actions: {
					$save: {
						$links: {
							$query: {
								$isHidden: true
							},
							$create: {
								$isHidden: true
							}
						}
					}
				}
			},
			$layout: {
				$items: [{
					$layoutType: "row",
					$items: [{
						$items: [{
							$title: "{@DocumentMainSectionTitle}",
							$category: "section",
							$layout: {
								$items: [{
									$bind: "description"
								}, {
									$bind: "isReadOnly"
								}, {
									$bind: "volume"
								}, {
									$bind: "owner"
								}]
							}
						}]
					}, {
						$items: [{
							$title: "{@DocumentOrganizationSectionTitle}",
							$category: "section",
							$layout: {
								$items: [{
									$bind: "tags"
								}, {
									$bind: "teams"
								}]
							}
						}]
					}]
				}, {
					$layoutType: "row",
					$items: [{
						$items: [{
							$title: "{@DocumentRelationsSectionTitle}",
							$category: "section",
							$layout: {
								$items: [{
									$bind: "className"
								}, {
									$bind: "x3Keys"
								}, {
									$bind: "representationName",
								}, {
									$bind: "endpoint",
								}]
							}
						}]
					}]
				}]
			}
		}
	}
};
