// syracuse-collaboration/lib/representations/msoffice/msoMailMergeTemplate.js
exports.representation = {
	$entityName: "msoWordTemplateDocument",
	$facets: {
		$query: {
			$prototype: {
				$links: {
					$query: {
						$isHidden: true
					},
					$create: {
						$isHidden: true
					},
					$delete: {
						$isHidden: true
					},
					$excel: {
						$isHidden: true
					},
					$print: {
						$isHidden: true
					},
					$wordmailmerge: {
						$isHidden: true
					},
					$wordreport: {
						$isHidden: true
					},
                    saveDocument: {
                        $title: "Save document to X3",
                        $type: "application/json;vnd.sage=syracuse",
                        $url: "/sdata/syracuse/collaboration/syracuse/msoWordTemplateDocuments/$template/$workingCopies?representation=msoMailMergeTemplate.$edit",
                        $method: "POST"
                    }
				},
			},
			$layout: {
                $items: [ ]
			}
		},
		$edit: {
			$prototype: {
				$links: {
					$query: {
						$isHidden: true
					},
					$print: {
						$isHidden: true
					},
					$wordreport: {
						$isHidden: true
					},
					$save: {
						$title: "Save to X3"
					}
				},
				$actions: {
					$save: {
						$links: {
							$query: {
								$isHidden: true
							},
							$create: {
								$isHidden: true
							}
						}
					}
				}
			},
			$layout: {
                $items: [{
                    $layoutType: "columns",
                    $layoutSubType: "50,50",
                    $items: [{
                    	$items: [{
	                        $title: "{@DocumentMainSectionTitle}",
	                        $category: "section",
	                        $layout: {
								$items: [{
									$bind: "description"
								}, {
									$bind: "isReadOnly"
								}, {
									$bind: "owner"
								}]
	                        }
                    	}]
					}, {
                    	$items: [{
	                        $title: "{@DocumentRelationsSectionTitle}",
	                        $category: "section",
	                        $layout: {
								$items: [{
									$bind: "templateType"
								}, {
									$bind: "templateClass"
								}, {
									$bind: "localeCode"
								}]
	                        }
                    	}]
					}]
				}, {
                    $layoutType: "columns",
                    $layoutSubType: "50,50",
                    $items: [{
                    	$items: [{
	                        $title: "{@DocumentInfoSectionTitle}",
	                        $category: "section",
	                        $layout: {
								$items: [{
									$bind: "documentType"
								}, {
									$bind: "documentDate"
								}, {
									$bind: "fileName"
								}]
	                        }
                    	}]
                    }, {
                    	$items: [{
		                    $title: "{@DocumentOrganizationSectionTitle}",
		                    $category: "section",
		                    $layout: {
								$items: [{
									$bind: "tags"
								}, {
									$bind: "teams"
								}]
		                    }
                    	}]
                    }]
				}]
			}
		}
	}
}
