// syracuse-collaboration/lib/representations/msoffice/msoMailMergeDSSelEntity.js
"use strict";

exports.representation = {
	$entityName: "msoMailMergeDSSel",
	resources: function() {
		var locale = require("syracuse-core/lib/locale");
		var globals = require('streamline/lib/globals');
		return locale.resources(module, globals.context && globals.context.sessionLocale)();
	},
	$facets: {
		$edit: {
			$prototype: {
				$links: {
					$query: {
						$isHidden: true
					},
					$delete: {
						$isHidden: true
					},
					$edit: {
						$isHidden: true
					},
					$create: {
						$isHidden: true
					},
					$print: {
						$isHidden: true
					},
					$excel: {
						$isHidden: true
					},
					$wordmailmerge: {
						$isHidden: true
					},
					$wordreport: {
						$isHidden: true
					},
				},
				$actions: {
					$save: {
						$title: "{@SaveAction}"
					}
				}
			},
			$layout: {
				$items: [{
					$bind: "endpoint"
				}, {
					$bind: "entity"
				}, {
					$bind: "representation"
				}, {
					$bind: "title"
				}, {
					$bind: "fetchCount"
				}]
			},
		},
	}
};
