// syracuse-collaboration/lib/representations/msoffice/msoMailMergeDSSel.js
"use strict";

exports.representation = {
	resources: function() {
		var locale = require("syracuse-core/lib/locale");
		var globals = require('streamline/lib/globals');
		return locale.resources(module, globals.context && globals.context.sessionLocale)();
	},
	$entityName: "msoMailMergeDSSel",
	$facets: {
		$query: {
			$prototype: {
				$links: {
					$query: {
						$isHidden: true
					},
					$delete: {
						$isHidden: true
					},
					$edit: {
						$isHidden: true
					},
					$create: {
						$isHidden: true
					},
					$print: {
						$isHidden: true
					},
					$excel: {
						$isHidden: true
					},
					$wordmailmerge: {
						$isHidden: true
					},
					$wordreport: {
						$isHidden: true
					},

					selectDS: {
						$title: "{@SelectDSAction}",
						$type: "application/json; vnd-sage=syracuse",
						$url: "/sdata/syracuse/collaboration/syracuse/msoMailMergeDSSels/$template/$workingCopies?representation=msoMailMergeDSSel.$edit",
						$method: "POST"
					},
					browseMenu: {
						$title: "{@BrowseMenuAction}",
						$type: "application/json; vnd-sage=syracuse",
						$url: "/sdata/syracuse/collaboration/syracuse/msoMailMergeDSSels/$template/$workingCopies?representation=msoMailMergeDSSelEntity.$edit",
						$method: "POST"
					},
					browseEntity: {
						$title: "{@BrowseEntityAction}",
						$type: "application/json; vnd-sage=syracuse",
						$url: "/sdata/syracuse/collaboration/syracuse/msoMailMergeDSSels/$template/$workingCopies?representation=msoMailMergeDSSelMenu.$edit",
						$method: "POST"
					}
				}
			},
			$layout: {
				$items: []
			}
		},
		$edit: {

			$prototype: {
				$links: {
					$query: {
						$isHidden: true
					},
					$delete: {
						$isHidden: true
					},
					$edit: {
						$isHidden: true
					},
					$create: {
						$isHidden: true
					},
					$print: {
						$isHidden: true
					},
					$excel: {
						$isHidden: true
					},
					$wordmailmerge: {
						$isHidden: true
					},
					$wordreport: {
						$isHidden: true
					},
				},
				$actions: {
					$save: {
						$title: "{@SaveAction}"
					}
				}

			},
			$layout: {
				$items: [{
					$bind: "datasource"
				}, {
					$bind: "title",
					$isEditMode: false
				}, {
					$bind: "endpoint",
					$isEditMode: false
				}, {
					$bind: "menuItem",
					$isEditMode: false
				}, {
					$bind: "entity",
					$isEditMode: false
				}, {
					$bind: "representation",
					$isEditMode: false
				}, {
					$bind: "fetchCount",
					$isEditMode: false
				}]
			}
		}
	}
};
