// syracuse-collaboration/lib/representations/msoffice/msoPptDocument.js
"use strict";
exports.representation = {
	$entityName: "document",
	$facets: {
		$query: {
			$prototype: {
				$links: {
					$query: {
						$isHidden: true
					},
					$create: {
						$isHidden: true
					},
					$delete: {
						$isHidden: true
					},
					$excel: {
						$isHidden: true
					},

					$print: {
						$isHidden: true
					},
					$wordmailmerge: {
						$isHidden: true
					},
					$wordreport: {
						$isHidden: true
					},
					saveNewDocument: {
						$title: "Save as document",
						$type: "application/json;vnd.sage=syracuse",
						$url: "/sdata/syracuse/collaboration/syracuse/documents/$template/$workingCopies?representation=msoPptDocument.$edit",
						$method: "POST"
					}
				},
			},
			$layout: {
				$items: []
			}
		},
		$edit: {
			$prototype: {
				$links: {
					$query: {
						$isHidden: true
					},
					$print: {
						$isHidden: true
					},
					$wordreport: {
						$isHidden: true
					}
				},
				$actions: {
					$save: {
						$links: {
							$query: {
								$isHidden: true
							},
							$create: {
								$isHidden: true
							}
						}
					}
				}
			},
			$layout: {
				$items: [{
					$layoutType: "stack",
					$items: [{
						$bind: "description"
					}, {
						$bind: "owner"
					}, {
						$bind: "volume",
					}, {
						$bind: "isReadOnly"
					}, {
						$bind: "documentDate",
						//						$isHidden: true
					}, {
						$bind: "teams"
					}, {
						$bind: "tags"
					}]
				}]
			}
		}
	}
}
