// syracuse-collaboration/lib/representations/msoffice/msoPptDocument.js
"use strict";
exports.representation = {
	resources: function() {
		var locale = require("syracuse-core/lib/locale");
		var globals = require('streamline/lib/globals');
		return locale.resources(module, globals.context && globals.context.sessionLocale)();
	},
	$entityName: "document",
	$facets: {
		$query: {
			$prototype: {
				$links: {
					$query: {
						$isHidden: true
					},
					$create: {
						$isHidden: true
					},
					$delete: {
						$isHidden: true
					},
					$excel: {
						$isHidden: true
					},

					$print: {
						$isHidden: true
					},
					$wordmailmerge: {
						$isHidden: true
					},
					$wordreport: {
						$isHidden: true
					},
					$excelreport: {
						$isHidden: true
					},
					saveNewDocument: {
						$title: "{@SaveActionNewDocument}",
						$type: "application/json;vnd.sage=syracuse",
						$url: "/sdata/syracuse/collaboration/syracuse/documents/$template/$workingCopies?representation=msoPptDocument.$edit",
						$method: "POST"
					}
				},
			},
			$layout: {
				$items: []
			}
		},
		$edit: {
			$prototype: {
				$links: {
					$query: {
						$isHidden: true
					},
					$print: {
						$isHidden: true
					},
					$wordreport: {
						$isHidden: true
					}
				},
				$actions: {
					$save: {
						$links: {
							$query: {
								$isHidden: true
							},
							$create: {
								$isHidden: true
							}
						}
					}
				}
			},
			$layout: {
				$items: [{
					$layoutType: "stack",
					$items: [{
						$bind: "description"
					}, {
						$bind: "owner"
					}, {
						$bind: "volume",
					}, {
						$bind: "isReadOnly"
					}, {
						$bind: "documentDate",
						//						$isHidden: true
					}, {
						$bind: "teams"
					}, {
						$bind: "tags"
					}]
				}]
			}
		}
	}
};
