// syracuse-collaboration/lib/representations/msoffice/msoWordTemplateDocument.js
exports.representation = {
	$entityName: "msoWordTemplateDocument",
	$facets: {
		$details: {
			$layout: {
                $items: [{
                    $layoutType: "columns",
                    $layoutSubType: "50,50",
                    $items: [{
                    	$items: [{
	                        $title: "{@DocumentMainSectionTitle}",
	                        $category: "section",
	                        $layout: {
								$items: [{
									$bind: "description"
								}, {
									$bind: "isReadOnly"
								}, {
									$bind: "content"
								}, {
									$bind: "owner"
								}]
	                        }
                    	}]
					}, {
                    	$items: [{
	                        $title: "{@DocumentRelationsSectionTitle}",
	                        $category: "section",
	                        $layout: {
								$items: [{
									$bind: "templateType"
								}, {
									$bind: "templateClass"
								}]
	                        }
                    	}]
					}]
				}, {
                    $layoutType: "columns",
                    $layoutSubType: "50,50",
                    $items: [{
                    	$items: [{
	                        $title: "{@DocumentInfoSectionTitle}",
	                        $category: "section",
	                        $layout: {
								$items: [{
									$bind: "documentType"
								}, {
									$bind: "documentDate"
								}, {
									$bind: "fileName"
								}]
	                        }
                    	}]
                    }, {
                    	$items: [{
		                    $title: "{@DocumentOrganizationSectionTitle}",
		                    $category: "section",
		                    $layout: {
								$items: [{
									$bind: "tags"
								}, {
									$bind: "teams"
								}]
		                    }
                    	}]
                    }]
				}]
			}
		},
		$edit: {
			$layout: {
				$copy: "$details"
			}
		},
        $query: {
            $layout: {
                $items: [{
                    $category: "section",
                    $layout: {
                        $items: [{
                            $bind: "$resources",
                            $layout: {
                                $items: [{
                                    $bind: "description"
                                }, {
                                	$bind: "documentDate"
                                }, {
                                	$bind: "fileName"
                                }, {
									$bind: "templateType"
								}, {
									$bind: "templateClass"
								}, {                                	
                                    $bind: "documentType"
                                }, {
                                    $bind: "isReadOnly"
                                }, {
                                    $bind: "content"
                                }, {
                                    $bind: "owner"
                                }]
                            }
                        }]
                    }
                }]
            }
        },
        $lookup: {
            $layout: {
                $items: [{
                    $category: "section",
                    $layout: {
                        $items: [{
                            $bind: "$resources",
                            $layout: {
                                $items: [{
                                    $bind: "description"
                                }, {
                                	$bind: "documentDate"
                                }, {
                                	$bind: "fileName"
                                }, {
                                    $bind: "owner"
                                }]
                            }
                        }]
                    }
                }]
            }
        }
	}
}
