// @sage/syracuse-lib/src/auth/bearer.ts
"use strict";

import { helpers } from '@sage/syracuse-core';
var tracer = helpers.debug.tracer("session.trace");
import * as config from 'config';
const locale = require('streamline-locale');
import * as checkUser from '../auth/checkUser';
import * as authHelper from '../auth/helpers';
import { changePasswordError } from '../auth/changePassword';
import * as oauth2 from './oauth2';
import * as adminHelpers from '../collaboration/helpers';
import * as jsxml from 'js-xml';
import * as url from 'url';

function unauthorized() {
	return authHelper.unauthorized('Basic realm=' + config.session.realm);
}

// When the access token for OAuth2 is already present in authentication header, OAuth2 authentication can be done in batch mode (for Web services)

export function authenticate(_, request, response, session) {
	var parsedUrl = url.parse(request.url, true);
	if (!parsedUrl.query || !parsedUrl.query.oauth2) throw unauthorized();
	var oauth2Server = parsedUrl.query.oauth2;

	var credentials = /^Bearer\s+(\S+)/.exec(request.headers.authorization);
	if (!(credentials && credentials[1])) throw unauthorized();

	var db = adminHelpers.AdminHelper.getCollaborationOrm(_);
	var server = db.fetchInstance(_, db.model.getEntity(_, "oauth2"), {
		jsonWhere: {
			batchAuthentication: true,
			name: oauth2Server
		}
	});
	if (!server) throw authHelper.error(404, "internal error: no OAuth2 server with batch authentication found");
	if (!server.active(_)) throw authHelper.error(403, "internal error: oauth2 server not active: " + server.name(_));

	if (server.name(_) !== 'sageid') {
		var json = oauth2.getUserInfo(_, credentials[1], server.dataRequestURL(_));
		tracer && tracer("Result of user name ", json);
		var userField = server.userField(_) || "user";
		var login = userField.split('.').reduce(function(obj, field) {
			var m = /^(\w+)\[(\d+)\]$/.exec(field);
			return m ? obj[m[1]][m[2]] : obj[field];
		}, json);
		var user = checkUser.fromLoginPage(_, request, 'oauth2', login, null, null, server);
	} else {
		var sageIdOAuth = (config.sage_id && config.sage_id.oauth) ? config.sage_id.oauth : {};
		var login = oauth2.sageIdTokenValidation(_, request, response, credentials[1]);
		var user = checkUser.fromLoginPage(_, request, 'sage-id', login, null);
	}

	tracer && tracer("User authenticated.");
	session.authData = {
		user: login,
		authorization: request.headers.authorization
	};
	return true;
};
