// @sage/syracuse-lib/src/auth/dispatcher.ts
"use strict";

import * as adminHelpers from '../collaboration/helpers';
const locale = require('streamline-locale');
import * as querystring from 'querystring';
import * as url from 'url';
import * as config from 'config';
import * as authHelper from '../auth/helpers';
import * as checkUser from '../auth/checkUser';
import * as loginPage from '../auth/loginPage';
import * as changePassword from '../auth/changePassword';
import * as forgetMePage from '../auth/forgetMePage';
import { apis } from '@sage/syracuse-core';

function badRequest(_, request, response, message) {
	response.writeHead(400, {
		"content-type": "text/plain",
	});
	response.end(message);
	return false;
}

function invalidAuthMethod(_, request, response, tok) {
	return badRequest(_, request, response, "invalid authentication method: " + tok);
}

function _ensureUserProfileLoaded(_, session) {
	session && session.getUserProfile(_);
	return true;
}

exports.ensureAuthenticated = function(_, request, response, params) {
	function filterUrlRedirect(url) { // only URL's which load the client (i. e. with main.html) or which invoke workflow actions ($service) should survive login page
		// re-enable unit tests with login
		return url.indexOf("/syracuse-main/html/main.html") === 0 || url.indexOf("/$service") > 0 || url === "/test-runner/lib/client/testClient.html";
	}

	var session = request.session;
	if (!session) throw new Error("internal error: no session");

	var httpStatus, location;
	try {
		// Dispatch to authenticate handlers.
		// These handlers should throw if authentication fails and return true 
		// (without sending any response) if it succeeds.

		// handle persistent cookie: check if there is a login cookie and if it's valid

		// APS manage forget me
		if (params && params.forgetme) {
			if (session.checkPersistentCookie(_, request, false)) {
				session.removeLoginToken(_, request);
			}
		}


		// if session is already authenticated, invoke renew hook if any, otherwise we are ok
		if (session.authData) {
			if (!session.authData.renew) return true;
			return session.authData.renew(_, request, response);
		}



		if (session.checkPersistentCookie(_, request, true)) return true;

		const healthApi = apis.get('syracuse-health');
		if (healthApi && !healthApi.acceptsRequest(_, request, response)) return;

		// if request is authenticated by a client-side certificate, handle it
		if (request.connection && request.connection.authorized) {
			var certAuth = authHelper.getAuthModule("certificate");
			// allow SSL with client authentication for normal users when authentication method "certificate" is switched off
			// if (!certAuth) return invalidAuthMethod(_, request, response, tok);
			if (certAuth) return certAuth.authenticate(_, request, response, request.session) && _ensureUserProfileLoaded(_, request.session);
		}

		// if request carries an authorization header, handle it
		if (request.headers.authorization) {
			var tok = request.headers.authorization.split(' ')[0].toLowerCase();
			var authModule = authHelper.getAuthModule(tok);
			if (!authModule) return invalidAuthMethod(_, request, response, tok);
			var isAuth = authModule.authenticate(_, request, response, request.session);
			if (isAuth) {
				if (params && params.keepConnected) session.setupPersistentCookie(_, request);
				_ensureUserProfileLoaded(_, request.session);
			}
			return isAuth;
		}
		if (session.loginCookie) {
			location = "/auth/forgetMe/page";
		} else {
			location = "/auth/login/page";
		}

		session.loginError = "";

	} catch (ex) {
		location = ex.$httpLocation || "/auth/login/page";
		session.loginError = ex.message;
	}

	if (!/^\/auth\//.test(request.url)) {

		session.authTargetUrl = request.url ? (filterUrlRedirect(request.url) ? request.url : '/') : '/';

		if (session.authTargetUrl.indexOf("device=phone") > -1) {
			// Mobile client uses only ajax calls, so we need to redirect to the main
			// html page doing the ajax request instead of the url catched by the dispatcher
			session.authTargetUrl = request.headers.referer;
		}

		// Tablet client 
		if (request.headers.referer && request.headers.referer.indexOf("syracuse-tablet") > -1) {
			session.authTargetUrl = request.headers.referer;
		}
	}


	authHelper.redirect(_, request, response, location);

	// we are not authenticated
	return false;
};

export const dispatcher = authHelper.dispatcher(2, {
	login: loginPage.dispatch,
	forgetMe: forgetMePage.dispatch,
	changePassword: changePassword.dispatch,
	oauth2: require('../auth/oauth2').dispatch,
	saml2: require('../auth/saml2').dispatch,
	'sage-id': require('../auth/sage-id').dispatch,
	userinfo: require('../auth/userInfo').dispatch
});
