// @sage/syracuse-lib/src/auth/loginPage.ts
"use strict";

const locale = require('streamline-locale');
import * as authHelper from '../auth/helpers';
import { helpers } from '@sage/syracuse-core';
import { apis } from '@sage/syracuse-core';
import * as config from 'config';

function genPage(_, request, response) {
	const healthApi = apis.get('syracuse-health');
	if (healthApi && !healthApi.acceptsRequest(_, request, response)) return;

	// check if all endpoint have only on product associated. If it's the case we return the product name in order to be displayed in the page else nothin
	var params = {};
	params['productName'] = authHelper.getProductName();

	['signInLabel', 'loginLabel', 'passwordLabel', 'keepConnectedLabel', 'dbGo', 'ldapGo', 'extTitle', 'oauth2Title', 'saml2Title', 'sageIdTitle', 'sageIdSignOn', 'sageIdRegister', 'sageIdRegisterExist'].forEach(function(label) {
		params[label] = locale.format(module, label);
	});

	var oauth2s = require('../auth/oauth2').getServerList(_);
	var saml2s = require('../auth/saml2').getServerList(_);

	// set visibility flags for the different authentication methods which do not provide several servers
	['basic', 'digest', 'ldap', 'sage-id'].forEach(function(method) {
		params[method + "Visibility"] = authHelper.isAllowed(method, true) ? 'visible' : 'hidden';
	});
	// for saml2 and oauth2 there must be servers
	params["saml2Visibility"] = (authHelper.isAllowed("saml2", true) && saml2s.length) ? 'visible' : 'hidden';
	params["oauth2Visibility"] = (authHelper.isAllowed("oauth2", true) && oauth2s.length) ? 'visible' : 'hidden';

	params.extVisibility = (params.oauth2Visibility === 'visible' || params.saml2Visibility === 'visible') ? 'visible' : 'hidden';
	params.passwordVisibility = (params.basicVisibility === 'visible' || params.digestVisibility === 'visible' || params.sageerpx3Visibility === 'visible' || params.ldapVisibility === 'visible') ? 'visible' : 'hidden';

	params.js$oauth2s = JSON.stringify(oauth2s);
	params.js$saml2s = JSON.stringify(saml2s);
	params.js$hasCookie = request.session.checkPersistentCookie(_, request, false);

	// set error message if last authentication failed
	if (request.session && request.session.loginError) {
		params.errorMessage = request.session.loginError;
		// display it only once
		request.session.loginError = "";
	} else {
		params.errorMessage = "";
	}

	authHelper.genPage(_, response, __dirname + "/../../public/auth/login.html", params);


}

function submit(_, request, response) {
	var params = request.readAll(_);
	if (helpers.http.parseHeaders(request.headers || {})["content-type"] === "application/json") params = JSON.parse(params);
	if (!require('../auth/dispatcher').ensureAuthenticated(_, request, response, params)) return;


	authHelper.redirect(_, request, response, request.session.authTargetUrl || '/', true);
}

export const dispatch = authHelper.dispatcher(3, {
	page: genPage,
	submit: submit,
});
