// @sage/syracuse-lib/src/collaboration/entities/msoffice/msoWordTemplateDocument.ts
"use strict";

var date = require('@sage/syracuse-core').types.date;
import { globals } from 'streamline-runtime';
import { AdminHelper as adminHelper } from '../../../collaboration/helpers';
const locale = require('streamline-locale');

var templatePresetValues = {
	"msoReportTemplate.$edit": {
		"templateType": "report",
		"templateClass": "{templateClass}",
		"localeCode": "{templateLocale}"
	},
	"msoMailMergeTemplate.$edit": {
		"templateType": "mailmerge",
		"templateClass": "{templateClass}",
		"localeCode": "{templateLocale}"
	}
};

function _getProperties(_, instance) {
	var prop = instance.content(_);
	return (prop && prop.fileExists(_) && prop.getProperties(_)) || {};
}

export const entity = {
	$titleTemplate: "Document template",
	$valueTemplate: "{description}",
	$descriptionTemplate: "Word template management",
	$listTitle: "List of templates",
	$canCreate: false,
	$helpPage: "Administration-reference_Word-templates-Documents",
	$properties: {
		code: {
			$title: "Code",
			$isMandatory: true,
			$linksToDetails: true,
			$isUnique: true,
			$isDisabled(_, instance) {
				return !instance.$created;
			}
		},
		description: {
			$title: "Description",
			$isMandatory: true,
			$isLocalized: true,
			$linksToDetails: true,
			$isUnique: true
		},
		documentType: {
			$title: "Document type",
			$isHidden: true,
			/*$compute(_, instance, propName) {
				return _getProperties(_, instance).contentType;
			},*/
			$isReadOnly: true
		},
		documentDate: {
			$title: "Upload date",
			$type: "date",
			/*$compute(_, instance, propName) {
				var upDate = _getProperties(_, instance).uploadDate;
				return upDate && date.fromJsDate(upDate);
			},*/
			$isReadOnly: true,
			$isNullable: true
		},
		fileName: {
			$title: "Filename",
			/*$compute(_, instance, propName) {
				return _getProperties(_, instance).fileName;
			},*/
			$isReadOnly: true
		},
		templateType: {
			$title: "Template type",
			$type: "string",
			$isDisabled: true,
			$isHidden: false,
			$enum: [{
				$title: "Report",
				$value: "report"
			}, {
				$title: "Mail merge",
				$value: "mailmerge"
			}],
			$default: "report"
		},
		templateClass: {
			$type: "string",
			$title: "Template class",
			$isHidden: false,
			$isDisabled: true
		},
		uri: {
			$title: "Uri",
			$isDisabled(_, instance) {
				return (instance.content(_) && instance.content(_).fileExists(_));
			},
			$isDefined(_, instance) {
				return false;
			}
		},
		isReadOnly: {
			$title: "Read only",
			$type: "boolean",
			$default: false
		},
		content: {
			$title: "Content",
			$type: "binary",
			$storage(_, instance) {
				return "db_file";
			},
			$propagate(_, instance, val) {
				if (!instance.description(_) && val && val.fileName) instance.description(_, val.fileName);
				instance.documentType(_, val.contentType);
				instance.documentDate(_, date.today());
				instance.fileName(_, val.fileName);
			}
		},
		templatePurpose: {
			$type: "string",
			$title: "Purpose",
			$isMandatory: false,
			$isDisabled: false,
			$lookup: {
				entity: "lookupTemplatePurposes",
				field: "name",
				parameters: "templateClass={templateClass}&templateType={templateType}"
			}
		},
		localeCode: {
			$title: "Locale code",
			$isMandatory: false,
			$isDisabled: true,
			$isHidden: false,
			$lookup: {
				entity: "localePreference",
				field: "name"
			}
		},
		cpy: {
			$type: "string",
			$isMandatory: false,
			$isDisabled(_, instance) {
				return !instance.endpoint(_);
			},
			$isHidden(_, instance) {
				return !instance.endpoint(_);
			},
			$lookup(_, instance) {
				if (!instance.endpoint(_)) return null;
				var baseUrl = instance.endpoint(_).getBaseUrl(_);
				console.log(baseUrl);
				return {
					$url: baseUrl + "/COMPANY?representation=COMPANY.$lookup",
					$result: "CPY"
				};
			},
			$control(_, instance, val) {
				if (!instance.endpoint(_) || !val) return;
				var db = instance.endpoint(_).getOrm(_);
				var cpy = db.fetchInstance(_, db.getEntity(_, "COMPANY", "$query"), {
					jsonWhere: {
						CPY: val
					}
				});
				if (!cpy) instance.$addError(locale.format(module, "companyMissing", val, instance.endpoint(_).description(_)), "cpy");
			}
		},
		leg: {
			$type: "string",
			$isMandatory: false,
			$isDisabled(_, instance) {
				return !instance.endpoint(_);
			},
			$isHidden(_, instance) {
				return !instance.endpoint(_);
			},
			$lookup(_, instance) {
				if (!instance.endpoint(_)) return null;
				var baseUrl = instance.endpoint(_).getBaseUrl(_);
				return {
					$url: baseUrl + "/ATABDIV?representation=ATABDIV.$lookup&where=(NUMTAB eq 909)",
					$result: "CODE"
				};
			},
			$control(_, instance, val) {
				if (!instance.endpoint(_) || !val) return;
				var db = instance.endpoint(_).getOrm(_);
				var leg = db.fetchInstance(_, db.getEntity(_, "ATABDIV", "$query"), {
					jsonWhere: {
						NUMTAB: 909,
						CODE: val
					}
				});
				if (!leg) instance.$addError(locale.format(module, "legislationMissing", val, instance.endpoint(_).description(_)), "leg");
			}
		},
		activ: {
			$type: "string",
			$isMandatory: false,
			$isDisabled(_, instance) {
				return !instance.endpoint(_);
			},
			$isHidden(_, instance) {
				return !instance.endpoint(_);
			},
			$lookup(_, instance) {
				if (!instance.endpoint(_)) return null;
				var baseUrl = instance.endpoint(_).getBaseUrl(_);
				return {
					$url: baseUrl + "/ACTIV?representation=ACTIV.$lookup",
					$result: "CODACT"
				};
			},
			$control(_, instance, val) {
				if (!instance.endpoint(_) || !val) return;
				var db = instance.endpoint(_).getOrm(_);
				var activ = db.fetchInstance(_, db.getEntity(_, "ACTIV", "$query"), {
					jsonWhere: {
						CODACT: val
					}
				});
				if (!activ) instance.$addError(locale.format(module, "activMissing", val, instance.endpoint(_).description(_)), "activ");
			}
		},
		application: {
			$isHidden: true
		},
		contract: {
			$isHidden: true
		},
		dataset: {
			$isHidden: true
		},

	},
	$relations: {
		teams: {
			$title: "Teams",
			$type: "teams",
			$inv: "templateDocuments"
		},
		owner: {
			$title: "Owner",
			$type: "user",
			$isMandatory: false
		},
		tags: {
			$title: "Tags",
			$type: "documentTags",
			$inv: "templateDocuments"
		},
		endpoint: {
			$title: "Endpoint",
			$type: "endPoint",
			$isMandatory: false,
			$propagate(_, instance, val) {
				if (val) {
					instance.application(_, val.applicationRef(_).application(_));
					instance.contract(_, val.applicationRef(_).contract(_));
					instance.dataset(_, val.dataset(_));
				} else {
					instance.application(_, "");
					instance.contract(_, "");
					instance.dataset(_, "");
				}
			},
			$lookupFilter: {
				protocol: "x3"
			}
		}
	},
	$init(_, instance) {
		var self = this;
		var up = globals.context && globals.context.session && globals.context.session.getUserProfile(_);
		up && instance.owner(_, up.user(_));

		var params = (((globals.context && globals.context.request && globals.context.request.context) || {}).parameters || {});
		if (params.representation) {
			var values = templatePresetValues[params.representation];
			if (values) {
				Object.keys(values).forEach_(_, function(_, key) {
					var value = values[key];
					var setter = instance[key];
					var val = value && value.replace(/\{(.*?)\}/g, function(match, p1) {
						return params[p1] || "";
					});
					if (val) instance[key](_, val);
				});
			}
		}
	}
};
