// syracuse-ui/lib/desktop/article/box.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Box = require('syracuse-ui/lib/common/article/box').Box;

function DesktopMenus(){

}

exports.DesktopMenus = helpers.defineClass(DesktopMenus, Box, {
    drawBox: function(){
        this.$skin = this.$item.$skin || (this.getArticle().$skin + "-" + this.$skinLevel); // this.$item.$category);
        this.$$body = $("<ul/>").css("display", "none");
        this.$$item = $("<nav/>").addClass(this.$skin);
        this.$$item.appendTo(this.$$container);
        this.$$body.addClass(this.$skin + "-body").appendTo(this.$$item);
        this.applyDesignMetaData(this.$item, false);
        if (this.$item.$width) {
            this.$$item.css("width", this.$item.$width);
        }
        if (this.$item.$css) {
            this.$$item.addClass(this.$item.$css);
        }
        this.menuItems = [];
        this.openBox(this.$item.$opened !== false);
    },
    toggleMore: function($$link){
        var self = this;
        if (!self.popupMore) {
            document.site.toggleClass(self.moreAddins.$$link, "s-close", false);
            self.popupMore = document.site.createPopup({
                page: self.getPage(),
                $$popup: self.moreAddins.$$body.toggle(true),
                position: {
                    my: "left top",
                    at: "left bottom",
                    of: self.moreAddins.$$linkLi
                    /*,
                     offset :"0 2"*/
                },
                closeHandler: function(){
                    document.site.toggleClass(self.moreAddins.$$link, "s-close", true);
                    self.popupMore = null;
                    return true;
                },
                boundarySelector: "#" + self.id
            });
        }
        else {
            self.popupMore.close();
        }
    },
    _appendMore: function($skinItem){
        var domMore = document.createElement("a");
        this.moreAddins = {
            $$linkLi: $("<li/>").addClass($skinItem + "-li").appendTo(this.$$body),
            $$link: $(domMore),
            $$body: $("<div/>").addClass($skinItem + "-more-body").appendTo(this.$$item).hide(),
            lastColCount: 0
        };
        domMore.className = $skinItem + "-more-btn s-close";
        domMore.setAttribute("data-s-more", "1");
        domMore.setAttribute("data-s-menu", this.id);
        domMore.setAttribute("data-s-article", this.getArticle().id);
        this.moreAddins.$$link.appendTo(this.moreAddins.$$linkLi).text("more");
    },
    addMenuItem: function($menu, record){
        if (!this.$item.$excludeBind || this.$item.$excludeBind.indexOf($menu.$bind) < 0) {
            var $$root = this.$$body;
            var $skinItem = this.$skin;
            var action = "appendTo";
            if (this.$item.$mainAction == $menu.$bind) {
                action = "prependTo";
                $skinItem += "-main";
            }
            if (this.$moreLimit < ++this.menuItemsCount) {
                if (!this.moreAddins) {
                    this._appendMore($skinItem);
                }
                if (this.moreAddins.lastColCount == 2) {
                    this.moreAddins.lastColCount = 0;
                }
                if (this.moreAddins.lastColCount == 0) {
                    this.moreAddins.$$lastCol = $("<ul/>").addClass(this.$skin + "-more-nav").appendTo(this.moreAddins.$$body);
                }
                this.moreAddins.lastColCount++;
                $$root = this.moreAddins.$$lastCol;
                $skinItem += "-more";
            }
            var menuItem = document.itemFactory.create({
                $bind: $menu.$bind,
                $skin: $skinItem + "-link",
                $category: $menu.$category || "link"
            }, this);
            var li = document.createElement("li");
            li.className = $skinItem + "-li";
            if (menuItem.$item.$css) {
                li.className += (" " + (menuItem.$item.$css));
            }
            (menuItem.$$liParent = menuItem.$$container = $(li))[action]($$root);
            menuItem.loadBox();
            menuItem.loaded = true;
            menuItem.setMenu($menu, record);
            this.menuItems.push(menuItem);
            return menuItem;
        }
    },
    addMenuItems: function($menus, $isAction){
        var self = this;
        Object.keys($menus).forEach(function($bind){
            var $menu = $menus[$bind];
            $menu.$bind = $bind;
            $menu.$isAction = $isAction;
            self.addMenuItem($menu);
        });
    },
    _renderLayoutContent: function(){
        //this.$$item.hide()=> if hideen show if menu in menuItem
        //this.$isHidden = true;
        this.menuItemsCount = 0;
        if (!this.$item.$columnLimit) {
            this.$moreLimit = this.$item.$moreLimit || 20;
        }
        if (!this.$item.$isBindDisabled) {
            this.getArticle().menusFacade.bindBox(this);
        }
        else {
            if (this.$item.$links) {
                this.addMenuItems(this.$item.$links);
            }
            if (this.$item.$actions) {
                this.addMenuItems(this.$item.$actions, true);
            }
        }
    },
    dispose: function(){
        if (this.menuItems) {
            var article = this.getArticle();
            if (article) {
                this.menuItems.forEach(function(menuItem){
                    article.removeItem(menuItem);
                });
                delete this.menuItems;
            }
        }
        delete this.onMenuClick;
        Box.prototype.dispose.call(this);
    }
});
