// syracuse-ui/lib/desktop/article/desktopSubMenu.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function DesktopSubMenu(){

}

exports.DesktopSubMenu = helpers.defineClass(DesktopSubMenu, null, {
    load: function(menus, $group, subMenuParent){
        var self = this;
        self.menus = menus;
        self.$subMenus = $group.$subMenus || {};
        
        var dom = document.createElement("li");
        dom.className = self.menus.$skin + "-li";
        var $$li = $(dom).appendTo(subMenuParent ? subMenuParent.$$group : self.menus.$$body);
        
        dom = document.createElement("ul");
        dom.className = (self.$subMenus.$skin || self.menus.$skin) + "-body";
        self.$$group = $(dom);
        
        if (self.$subMenus.$isPopup) {
            dom = document.createElement("a");
            dom.className = self.menus.$skin + "-sub-picker";
            self.$$link = $(dom).appendTo($$li);
            
            dom = document.createElement("div");
            dom.className = self.menus.$skin + "-popup-body";
            self.$$popupBody = $(dom).append(self.$$group).hide();
            self.$$link.bind("click", function(){
                if (!self._popupMenus) {
                    document.site.toggleClass(self.$$link, "s-open", true);
                    self._popupMenus = self.menus.boxParent.openDialog({
                        $dialogMode: "popup",
                        content: self,
                        $$dialog: self.$$popupBody.toggle(true),
                        position: {
                            my: "right top",
                            at: "right bottom",
                            of: self.$$link
                        },
                        onClose: function(){
                            document.site.toggleClass(self.$$link, "s-open", false);
                            self._popupMenus = null;
                        }
                    });
                }
                else {
                    self._popupMenus.close();
                }
                return false;
            });
        }
        else {
            dom = document.createElement("a");
            dom.className = self.menus.$skin + "-sub-root";
            self.$$link = $(dom).appendTo($$li);
            self.$$group.appendTo($$li)
        }
        if (self.menus.$item.$css) {
            self.$$link.addClass(self.menus.$item.$css).text($group.$title);
        }
        self.$$link.text($group.$title);
        
    },
    onChildMenuItemClick: function(){
        if (this._popupMenus) {
            this._popupMenus.close();
        }
    },
    dispose: function(){
        if (this.$$link && this.$subMenus.$isPopup) {
            this.$$link.unbind();
            this.$$group.empty();
        }
        delete this.menus;
    }
});
