// syracuse-ui/lib/desktop/article/desktopSectionBlock.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Box = require('syracuse-ui/lib/common/article/box').Box;

var _localize = {
    sectionTitle: "Section",
    blockTitle: "block"
};
function DesktopSectionBlock(){

}

exports.DesktopSectionBlock = helpers.defineClass(DesktopSectionBlock, Box, {
    getDefaultTitle: function(){
        return _localize[this.$authoringLevel + "Title"];
    },
    switchCategory: function($newCategory){
        var self = this;
        var $prevAuthoringLevel = self.$authoringLevel;
        self.$authoringLevel = self.$item.$category = $newCategory;
        self._ensureSkin(self.$authoringLevel);
        ["$$item", "$$body", "$$header", "$$title"].forEach(function(prop){
            if (self[prop]) {
                var dom = self[prop][0];
                dom.className = dom.className.replace(new RegExp($prevAuthoringLevel, "g"), self.$authoringLevel);
            }
        });
        if (self.authoringNode) {
            self.authoringNode.$authoringLevel = $newCategory;
        }
    },
    loadBox: function(){
        this._ensureSkin(this.$authoringLevel);
        Box.prototype.loadBox.call(this);
    }
});
