// syracuse-ui/lib/desktop/article/page.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Page = require('syracuse-ui/lib/common/article/page').Page;

function DesktopPage(){
}

exports.DesktopPage = helpers.defineClass(DesktopPage, Page, {
    loadBox: function(initData){
        switch (this.$pageCategory) {
            case "dashboard":
                this.$authoringSubType = "dashboard";
                this.$isSearchFacetEnabled = false;
                this.$autoFetch = false;
                break;
            case "portlet":
                if (this.$facet == "$demo") {
                    this.$autoFetch = false;
                }
                this.$isSearchFacetEnabled = false;
                this.$item.$isTitleHidden = true;
                break;
            default:
                this.$isSearchFacetEnabled = true;
                break;
        }
        Page.prototype.loadBox.call(this, initData);
    },
    
    appendArticleMenus: function(){
        if (this.$pageCategory != "dashboard") {
            Page.prototype.appendArticleMenus.call(this, (this.$pageCategory == "portlet") ? "prependTo" : null);
        }
    },
    drawUserProfile: function(userProfile){
        if (document.site.$item.$isUserProfileQuickViewEnabled !== false) {
            var dom;
            if (!this.$$headerUserProfile) {
                dom = document.createElement("div");
                dom.className = this.$skin + "-head-user-profile";
                this.$$headerUserProfile = $(dom).appendTo(this.$$header);
            }
            else {
                this.$$headerUserProfile.empty();
            }
            if (userProfile) {
                if (userProfile.selectedRole) {
                    dom = document.createElement("label");
                    dom.className = this.$skin + "-head-user-role";
                    this.$$headerUserProfile.append($(dom).text(userProfile.selectedRole.description || ""));
                }
                if (userProfile.selectedEndpoint) {
                    dom = document.createElement("label");
                    dom.className = this.$skin + "-head-user-endpoint";
                    this.$$headerUserProfile.append($(dom).text(userProfile.selectedEndpoint.description || ""));
                }
                userProfile.selectedRole && userProfile.selectedRole.description
            }
        }
    },
    appendHeader: function($$container){
        Page.prototype.appendHeader.call(this, $$container);
        if (this.$prototype.$description) {
            this.renderDescription(this.$prototype.$description);
        }
        this.$$header.append(this.$$headerLeft = $("<div/>").addClass(this.$skin + "-head-left").append(this.$$header.children()));
    },
    drawBox: function(){
        var dom;
        this.$item.$title = this.$item.$title || "{$title}";
        if (this.$item.$skin) {
            console.log("page : draw box $skin is undefined");
        }
        this.$skin = this.$item.$skin || this.$layoutOptions.$skin;
        dom = document.createElement("article");
        dom.className = this.$skin;
        this.$$item = $(dom).appendTo(this.$$container);
        
        dom = document.createElement("div");
        dom.className = this.$skin + "-data";
        this.$$dataSlot = $(dom).appendTo(this.$$item);
        if (!this.$item.$isTitleHidden) {
            this.appendHeader(this.$$dataSlot);
        }
        
        dom = document.createElement("div");
        dom.className = this.$skin + "-body";
        this.$$body = $(dom).appendTo(this.$$dataSlot);
        
        this._renderLayoutContent(this.$item);
        this.appendArticleMenus();
        if (this.$$menus && this.$isSearchFacetEnabled) {
            document.itemFactory.load(this.$$menus, {
                $category: "searchFacets",
                $prototype: {},
                $bind: "$searchFacets"
            }, this);
        }
    },
    
    applyChange: function(newData){
        Page.prototype.applyChange.call(this, newData);
        if (this.dialog) {
            this.dialog.center();
        }
    },
    dispose: function(){
        delete this.dialog;
        Page.prototype.dispose.call(this);
    }
});
