// syracuse-ui/lib/desktop/article/page.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Page = require('syracuse-ui/lib/common/article/page').Page;

function DesktopPage(){
}

exports.DesktopPage = helpers.defineClass(DesktopPage, Page, {
    _initializeBox: function(){
        switch (this.$pageCategory) {
            case "dashboard":
                this.$authoringSubType = "dashboard";
                this.$autoFetch = false;
                break;
            case "portlet":
                if (this.$facet == "$demo") {
                    this.$autoFetch = false;
                }
                this.$item.$isTitleHidden = true;
                break;
        }
        Page.prototype._initializeBox.call(this);
    },
    appendArticleMenus: function(){
        if (this.$pageCategory != "dashboard") {
            Page.prototype.appendArticleMenus.call(this, (this.$pageCategory == "portlet") ? "prependTo" : null);
        }
    },
    drawUserProfile: function(userProfile){
        if (document.site.$item.$isUserProfileQuickViewEnabled !== false) {
            var dom;
            if (!this.$$headerUserProfile) {
                dom = document.createElement("div");
                dom.className = this.$skin + "-head-user-profile";
                this.$$headerUserProfile = $(dom).appendTo(this.$$header);
            }
            else {
                this.$$headerUserProfile.empty();
            }
            if (userProfile) {
                if (userProfile.selectedRole) {
                    dom = document.createElement("label");
                    dom.className = this.$skin + "-head-user-role";
                    this.$$headerUserProfile.append($(dom).text(userProfile.selectedRole.description || ""));
                }
                if (userProfile.selectedEndpoint) {
                    dom = document.createElement("label");
                    dom.className = this.$skin + "-head-user-endpoint";
                    this.$$headerUserProfile.append($(dom).text(userProfile.selectedEndpoint.description || ""));
                }
                userProfile.selectedRole && userProfile.selectedRole.description
            }
        }
    },
    appendHeader: function($$container){
        Page.prototype.appendHeader.call(this, $$container);
        if (this.$prototype.$description) {
            this.renderDescription(this.$prototype.$description);
        }
        this.$$header.append(this.$$headerLeft = $("<div/>").addClass(this.$skin + "-head-left").append(this.$$header.children()));
    },
    drawBox: function(){
        var dom;
        this.$item.$title = this.$item.$title || "{$title}";
        /*if (this.$item.$skin) {
         debugger;
         }*/
        this.$skin = this.$item.$skin || document.site.$item.$layoutOptions[this.$pageCategory].$skin;
        dom = document.createElement("article");
        dom.className = this.$skin;
        this.$$item = $(dom).appendTo(this.$$container);
        
        dom = document.createElement("div");
        dom.className = this.$skin + "-data";
        var $$data = $(dom).appendTo(this.$$item);
        if (!this.$item.$isTitleHidden) {
            this.appendHeader($$data);
        }
        
        dom = document.createElement("div");
        dom.className = this.$skin + "-body";
        this.$$body = $(dom).appendTo($$data);
        
        this._renderLayoutContent(this.$item);
        this.appendArticleMenus();
        if (this.$$menus) {
            document.itemFactory.load(this.$$menus, {
                $category: "searchFacets",
                $prototype: {},
                $bind: "$searchFacets"
            }, this);
        }
        if (this.$displayTarget == "modal") {
            this._appendButton("s-dialog-page-close", "onClose");
            this.$$header.addClass("s-dialog-page-head");
            $$data.addClass("s-dialog-page");
            this.$cssDialog = "s-dialog-page-container";
            this.$$item.addClass("s-dialog-page").hide();
            this.$$body.css({
                overflow: "auto",
                height: 0
            });
            document.site.applyDialogDecorator(this);
            this.$$item.show();
        }
    },
    onDialogResize: function(){
        var body = this.$$body[0];
        var $$dialog = document.site.dialogAdapter.$$dialog;
        var root = $$dialog.parent()[0];
        var top = 0;
        var start = body;
        while (start && start != root) {
            top += start.offsetTop;
            start = start.offsetParent;
        }
        body.style.height = ($$dialog.height() - top) + "px";
    },
    applyChange: function(newData){
        Page.prototype.applyChange.call(this, newData);
        if (document.site.dialogAdapter) {
            document.site.dialogAdapter.center();
        }
    },
    dispose: function(){
        delete this.$layoutOptions;
        if (document.site.dialogAdapter) {
            document.site.removeDialog();
        }
        Page.prototype.dispose.call(this);
    }
});
