// syracuse-x3/lib/fusion/AdapiObservable.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var EventEmitter = require('events').EventEmitter;

var AdapiObservable = exports.AdapiObservable = helpers.defineClass(function(adapiClient) {
	EventEmitter.call(this);
	this.adapiClient = adapiClient;
	this.jsonType = null;
	this.notified = false;
	//if (adapiClient != null) addObserver(adapiClient);
}, EventEmitter, {
	addObserver: function(o) {
		// TODO
	},
	compareTo: function(o) {
		return this.getAdapiObservableId().compareTo(o.getAdapiObservableId());
	},
	dispose: function() {
		deleteObservers();
	},
	forceChange: function() {
		if (false && this.notified) {
			return;
		}
		if (this.jsonType != null) {
			// send only when jsonNode Type is defined
			// TODO
			//this.setChanged();
			//this.notifyObservers();
		}
		this.notified = true;
	},
	getAdapiObservableId: function() {
		return this.class.name + hashCode();
	},
	razModification: function() {
		this.notified = false;
	},
	setJsonType: function(aType) { // change jsontype and
		// force change
		this.jsonType = aType;
		this.forceChange();
	}
	//toJSON: function();
});
