// syracuse-x3/lib/fusion/AdapiStatusBar.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

var AdapiStatusIcon = exports.AdapiStatusIcon = helpers.defineClass(function(aImgId, aActionId, aVisible) {
	this.imgId = aImgId;
	this.actionId = aActionId;
	this.visible = aVisible;
}, null, {
	toJson: function() {
		// this.actionId = 0 pour les icone supprimés
		var ico = new JSONArray();
		ico.put(this.visible ? this.actionId : 0);
		ico.put(this.imgId);
		return ico;
	},
	update: function(aActionId, aVisible) {
		this.actionId = aActionId;
		this.visible = aVisible;
	}
});

var AdapiStatusBar = exports.AdapiStatusBar = helpers.defineClass(function() {
	this.iconsList = new HashMap(Integer, AdapiStatusIcon)();
	this.text = [];

}, null, {
	addIconDef: function(aNumtext, aText, aImgId, aActionId) {
		if (aImgId == 0) {
			setText(aNumtext, aText);
		} else {
			var wVisible = aImgId > 0;
			if (!wVisible) {
				aImgId = aImgId * -1;
			}
			var key = new Integer(aImgId);
			var wStatusIcon = this.iconsList.get(key);
			if (wStatusIcon == null) {
				this.iconsList.put(key, new AdapiStatusIcon(aImgId, aActionId, wVisible));
			} else {
				wStatusIcon.update(aActionId, wVisible);
			}
		}
	},
	setText: function(aNumText, aText) {
		if (aNumText < 1 || aNumText > 2) {
			aNumText = 1; // protection
		}
		if (aText != null && aText.length() > 0) {
			this.text[aNumText - 1] = aText;
		} else {
			this.text[aNumText - 1] = null;
		}
	},
	toJson: function() {
		var statusbar = new JSONObject();
		var text = new JSONArray();
		for (var t in this.text) {
			if (t == null) {
				text.put("");
			} else {
				text.put(t);
			}
		}
		if (text.length() != 0) {
			statusbar.put('t', text);
		}
		var ico = new JSONArray();
		for (var item in this.iconsList.values()) {
			ico.put(item.toJson());
		}
		if (ico.length() != 0) {
			statusbar.put('ico', ico);
		}
		return statusbar;
	}
});
