// syracuse-x3/lib/fusion/CAdapiMessages.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var CReplyMessages = Array; //require('syracuse-x3/lib/fusion/CReplyMessages').CReplyMessages;

var CAdapiMessages = exports.CAdapiMessages = helpers.defineClass(function(clientObs) {
	CReplyMessages.call(this);
	this.clientObs = clientObs;
}, CReplyMessages, {
	addMessage: function(aType, aMessage) {
		addElement(new CAdapiMessage(aType, clientObs, aMessage, false));
	},
	getMess: function(aIdx) {
		var wMess;
		try {
			wMess = elementAt(aIdx).mess;
		} catch (e) {
			wMess = "message idx=[" + aIdx + "] does not exist.";
		}
		return wMess;
	},
	getTypMess: function(aIdx) {
		var wTypMess;
		try {
			wTypMess = elementAt(aIdx).typ;
		} catch (e) {
			wTypMess = -1;
		}
		return wTypMess;
	},
	hasErrorMessage: function() {
		var wEnum = elements();
		while (wEnum.hasMoreElements()) {
			if (wEnum.nextElement().isErrorMessage()) {
				return true;
			}
		}
		return false;
	},
	hasMessages: function() {
		return this.length > 0;
	},
	hasStrongMessage: function() {
		var wEnum = elements();
		while (wEnum.hasMoreElements()) {
			if (wEnum.nextElement().isStrongMess()) {
				return true;
			}
		}
		return false;
	}
});
