// syracuse-x3/lib/fusion/CAdapiInstance.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var AdapiObservable = require('syracuse-x3/lib/fusion/AdapiObservable').AdapiObservable;

var EAdapiInstanceKind = {
	TARGET: 0,
	PREVIOUS: 1
};

var CAdapiInstance = exports.CAdapiInstance = helpers.defineClass(function(cli, aKind, aIst, format, aValue, modCmd) {
	AdapiObservable.call(this, cli);
	this.nl = -1;
	this.exception = null;
	if (cli == null) {
		throw new RuntimeException("Unexpected null FusionClient");
	}
	this.ist = aIst;
	this.kind = aKind;
	this.roughValue = null;
	this.modCommande = modCmd;
	this.format = format == null || format.isEmpty() ? null : format;
	this.value = aValue;
	var adapi = cli.findModifiable(aIst);
	if (adapi != null && adapi instanceof CAdapiEntity) {
		this.adapiEntity = adapi;
	}
	if (this.adapiEntity == null) {
		throw new RuntimeException("Entity " + aIst.toStringDescr() + " not found\nPlease check/update X3 Window description");
	}
	// !! En dernier car update du client et this.adapiEntity doit être valorisee
	if (aKind == EAdapiInstanceKind.PREVIOUS) {
		setJsonType('PREVIOUS');

	} else {
		setJsonType('TARGET');
	}
}, AdapiObservable, {
	getNl: function() {
		if (this.nl != -1) {
			return this.nl;
		} else {
			var v = this.adapiEntity instanceof CAdapiVariable ? this.adapiEntity : null;
			if (v != null && !v.isDisposed()) {
				var line = v.srcLine;
				if (line != null && line.dataSrc.isBlocGrid()) {
					return line.nodeId;
				}
			}
		}
		return -1;
	},
	getXid: function() {
		var fullId = this.ist.screenAlphaIdFusion;
		if (!this.adapiEntity.isDisposed() && !(this.adapiEntity instanceof CAdapiScreenDummy)) {
			fullId += this.ist.blocAlphaId;
			if (this.adapiEntity.entityTyp == 'FIELD') {
				fullId += this.ist.field;
			}
		}
		return fullId;
	},
	hasException: function() {
		return this.exception != null;
	},
	hasFormat: function() {
		return this.format != null;
	},
	setException: function(e) {
		this.exception = e;
	},
	setNl: function(nl) {
		this.nl = nl;
	},
	setRoughValue: function(aValue) {
		this.roughValue = aValue;
	},
	toJson: function() {
		var object = new JSONObject();
		if (this.ist.ist.winAlphaId == 1) {
			// target portal
			object.put('type', 'portal');
		} else {
			object.put('type', 'ist');
			var ist = new JSONObject();
			ist.put('xid', this.getXid());
			if (this.getNl() != -1) {
				ist.put('nl', this.getNl());

			}
			if (this.format != null) {
				ist.put('fmt', this.format);
			}
			ist.put('win', this.ist.winAlphaId);
			ist.put('edit', !this.modeCommande);
			// Valeur pour champ en mode edit - TODO Vérifier si ok
			if (!this.modeCommande && this.value != null) {
				ist.put('v', this.value);
				ist.put('raw', this.roughtValue);
			}
			object.put('ist', ist);
			ist.put('ctrlFailed', hasException());
			if (hasException()) {
				var diags = new CDiagnoses();
				var diag = new CDiagnosis();
				diag.setMessage(this.exception.message);
				diags.addDiagnosis(diag);
				ist.put('$diagnoses', diags.toJSONArray());
			}
		}
		return object;
	}
});
CAdapiInstance.newprevious = function(aClient, aIst, aFormatString, aValue, modCommande) {
	return new CAdapiInstance(aClient, EAdapiInstanceKind.PREVIOUS, aIst, aFormatString, aValue, modCommande);
}

CAdapiInstance.newtarget = function(aClient, aIst, aFormatString, aValue, modCommande) {
	return new CAdapiInstance(aClient, EAdapiInstanceKind.TARGET, aIst, aFormatString, aValue, modCommande);
}
