// syracuse-x3/lib/fusion/CAdapiScreen.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var CAdapiEntity = require('syracuse-x3/lib/fusion/CAdapiEntity').CAdapiEntity;
var CAdapiListOfDataSrcs = require('syracuse-x3/lib/fusion/CAdapiListOfDataSrcs').CAdapiListOfDataSrcs;


var SCREEN_ID_BUTTONS = "_BUTTONS";
var SCREEN_ID_TABTITLES = "_TABTITLE";
var CAdapiScreen = exports.CAdapiScreen = helpers.defineClass(function(requesterWindow, idxbaseOne, aScreenId, aListOfEntity) {
	CAdapiEntity.call(this, requesterWindow && requesterWindow.client, aScreenId, idxbaseOne, 'SCREEN');
	this.requesterWindow = requesterWindow;
	this.listOfEntity = aListOfEntity !== undefined ? aListOfEntity : new CAdapiListOfDataSrcs(requesterWindow);
	// En dernier
	this.setJsonType('ENTITY');
}, CAdapiEntity, {
	getAdxId: function() {
		if (this.requesterWindow == null) {
			return null;
		}
		return "" + this.requesterWindow.getAdxId() + ',' + this.entityIdxAlpha;
	},
	getJsonKey: function() {
		return this.getAdapiObservableId().substring(1, 2);
	},
	isScreenButtons: function() {
		return this.entityIndexBaseOne == CAdapiScreen.SCREEN_BUTTONS_NUM_ZERO_BASED + 1;
	},
	isScreenLeftList: function() {
		return this.entityIndexBaseOne == CAdapiScreen.SCREEN_LEFTLIST_NUM_ZERO_BASED + 1;
	},
	isScreenTabbed: function() {
		return this.entityIndexBaseOne < CAdapiScreen.SCREEN_BUTTONS_NUM_ZERO_BASED + 1;
	},
	isScreenTabTitle: function() {
		return this.entityIndexBaseOne == CAdapiScreen.SCREEN_TABTITLES_NUM_ZERO_BASED + 1;
	},
	pushDataSrc: function(aId, aDataSrc) {
		this.listOfEntity.pushEntity(aId, aDataSrc);
	},
	toJSON: function() {
		var object = {};
		// Modif Sreen
		if (this.hasModifTitle()) {
			object.tit = this.title;
			if (this.titleSty != null) {
				object.stytit = this.titleSty;
			}
		}
		if (this.hasModifStyle()) {
			object.sty = this.style;
		}
		return object;
	},
	traverseAll: function(aTraverser) {
		return this.listOfEntity.traverseAll(aTraverser);
	},
	traverseChilds: function(aTraverser) {
		return this.listOfEntity.traverseChilds(aTraverser);
	},
	traverseSettable: function(aStructIst, aAdapiSetter) {
		return this.listOfEntity.traverseSettable(aStructIst, aAdapiSetter);
	}
});


CAdapiScreen.screenZeroBaseIdxToWhatScreen = function(aIdx) {
	if (aIdx == CAdapiScreen.SCREEN_LEFTLIST_NUM_ZERO_BASED) {
		return "33|SCREEN_LEFTLIST(34,AH)";
	} else if (aIdx == CAdapiScreen.SCREEN_BUTTONS_NUM_ZERO_BASED) {
		return "32|SCREEN_BUTTONS(33,AG)";
	} else if (aIdx == CAdapiScreen.SCREEN_TABTITLES_NUM_ZERO_BASED) {
		return "34|SCREEN_TABTITLES(35,AI)";
	} else {
		return aIdx.toString();
	}
}
CAdapiScreen.SCREEN_BUTTONS_NUM_ZERO_BASED = 32; // index base "one" = 33
CAdapiScreen.SCREEN_LEFTLIST_NUM_ZERO_BASED = 33; // index base "one" =
CAdapiScreen.SCREEN_TABTITLES_NUM_ZERO_BASED = 34; // index base "one" =
