// syracuse-x3/lib/fusion/CAdapiLeftListLineNode.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var CAdapiLeftListLine = require('syracuse-x3/lib/fusion/CAdapiLeftListLine').CAdapiLeftListLine;


var CAdapiLeftListLineNode = exports.CAdapiLeftListLineNode = helpers.defineClass(function(aStackedWindow, aLeftList, aId, aNbVariables, idx) {
	CAdapiLeftListLine.call(this, aStackedWindow, aLeftList, aId, aNbVariables, idx);
	this.subLeftListNodes = [];
}, CAdapiLeftListLine, {
	attachLefListNode: function(aLeftListNode) {
		this.subLeftListNodes.push(aLeftListNode);
	},
	clone: function() {
		var idx = this.dataSrc.size();
		var wClone = new CAdapiLeftListLineNode(this.adapiRequester, this.dataSrc, this.id, this.vars.length, idx);
		cloneVars(wClone);
		return wClone;
	},
	hasSubNode: function() {
		return this.subLeftListNodes.length > 0;
	},
	setColumns: function(aIdx, aText, aClf) {
		setZeroBaseIdx(aIdx);
		setCle(aClf);
		var v = this.getVar(0);
		if (v != null) {
			v.setUnModifiableServerValue(new CStructResu(CStructResu.X3TYP_STRU, aText));
		}
	},
	traverseChilds: function(aTraverser) {
		var wTraversingState = ITraverser.CONTINUE_TRAVERSING;
		if (hasSubNode()) {
			wTraversingState = traverseSubNodes(aTraverser);
		}
		if (wTraversingState == ITraverser.CONTINUE_TRAVERSING) {
			CAdapiLeftListLine.prototype.traverseChilds.call(this,aTraverser);
		}
		return wTraversingState;
	},
	traverseNodes: function(aTraverser, aWithSubNodes) {

		var wTraversingState = ITraverser.CONTINUE_TRAVERSING;
		var wLeftListNode;
		var wMax = this.subLeftListNodes.length;
		var wI = 0;
		while (wI < wMax) {
			wLeftListNode = this.subLeftListNodes[wI];
			wTraversingState = aTraverser.doWhenTraverse(wLeftListNode, wI);
			if (wTraversingState != ITraverser.STOP_TRAVERSING && aWithSubNodes && wLeftListNode.hasSubNode()) {
				wTraversingState = wLeftListNode.traverseNodes(aTraverser, aWithSubNodes);
			}
			if (wTraversingState == ITraverser.STOP_TRAVERSING) {
				return ITraverser.STOP_TRAVERSING;
			}
			wI++;
		}
		return wTraversingState;
	},
	traverseSubNodes: function(aTraverser) {
		return traverseNodes(aTraverser, false);
	}
});
