// syracuse-x3/lib/fusion/CAdapiListOfScreens.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var CAdapiScreen = require('syracuse-x3/lib/fusion/CAdapiScreen').CAdapiScreen;
var CAdapiListOfEntity = require('syracuse-x3/lib/fusion/CAdapiListOfEntity').CAdapiListOfEntity;
var util = require('util');

var CAdapiScreenDummy = exports.CAdapiScreenDummy = helpers.defineClass(function(aAdapiRequesterWindow, aEntityIndexbaseOne, aScreenId) {
	CAdapiScreen.call(this, aAdapiRequesterWindow, aEntityIndexbaseOne, aScreenId, null);
}, CAdapiScreen, {

	traverseChilds: function(aTraverser) {
		return this.traverseChilds(aTraverser);
	},
	traverseSettable: function(aStructIst, aAdapiSetter) {
		return aAdapiSetter.setAdapiModifiable(aStructIst, this);
	}
});

var CAdapiListOfScreens = exports.CAdapiListOfScreens = helpers.defineClass(function(aAdapiRequesterWindow) {
	CAdapiListOfEntity.call(this, aAdapiRequesterWindow);
	this.screenLeftList = null;
	this.screenButtons = new CAdapiScreenDummy(aAdapiRequesterWindow, CAdapiScreen.SCREEN_BUTTONS_NUM_ZERO_BASED + 1, CAdapiScreen.SCREEN_ID_BUTTONS);
	this.screenTabTitles = new CAdapiScreenDummy(aAdapiRequesterWindow, CAdapiScreen.SCREEN_TABTITLES_NUM_ZERO_BASED + 1, CAdapiScreen.SCREEN_ID_TABTITLES);
}, CAdapiListOfEntity, {
	get: function(aIdx) {
		if (aIdx < CAdapiScreen.SCREEN_BUTTONS_NUM_ZERO_BASED && this.size() > aIdx) {
			return CAdapiListOfEntity.prototype.get.call(this,aIdx);
		} else if (aIdx == CAdapiScreen.SCREEN_BUTTONS_NUM_ZERO_BASED) //32 => index base "one" = 33 => AG
		{
			return this.screenButtons;
		} else if (aIdx == CAdapiScreen.SCREEN_LEFTLIST_NUM_ZERO_BASED) //33 => index base "one" = 34 => AH
		{
			return this.screenLeftList;
		} else if (aIdx == CAdapiScreen.SCREEN_TABTITLES_NUM_ZERO_BASED) //34 => index base "one" = 35 => AI
		{
			return this.screenTabTitles;
		} else {
			return null;
		}
	},
	pushScreenLeftList: function(aAdapiScreen) {
		this.screenLeftList = aAdapiScreen;
		this.push(aAdapiScreen.id, aAdapiScreen);
	},
	traverseAll: function(aTraverser) {
		var wMax = size();
		var wScreen;
		var wI = 0;
		while (wI < wMax) {
			wScreen = this.get(wI);
			if (aTraverser.doWhenTraverse(wScreen, wI) == true) {
				return true;
			}
			if (wScreen.traverseAll(aTraverser) == true) {
				return true;
			}
			wI++;
		}
		return false;
	},
	traverseChilds: function(aTraverser) {
		var wMax = size();
		var wEntry;
		var wI = 0;
		while (wI < wMax) {
			wEntry = this.get(wI);
			if (aTraverser.doWhenTraverse(wEntry, wI) == true) {
				return true;
			}
			wI++;
		}
		return false;
	},
	traverseSettable: function(aStructIst, aAdapiSetter) {
		var wI = aStructIst.screenZeroBase;
		var wAdapiScreen = this.get(wI);
		if (wAdapiScreen == null) {
			var wMess = this.adapiClient.adapiResources.getMessage(CAdapiResources.KEYMESS_ERROR_SCRN_UNKNOWN);
			wMess = CStringFacilities.sprintf(wMess, CAdapiScreen.screenZeroBaseIdxToWhatScreen(wI), size().toString(), this.adapiRequesterWindow.id);
			aAdapiSetter.setLogicalErrMess(wMess);
			return false;
		}
		if (!aStructIst.isScreenIst()) {
			return wAdapiScreen.traverseSettable(aStructIst, aAdapiSetter);
		} else {
			return aAdapiSetter.setAdapiModifiable(aStructIst, wAdapiScreen);
		}
	}
});
