// syracuse-x3/lib/fusion/CAdapiListOfScreens.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var CAdapiScreen = require('syracuse-x3/lib/fusion/CAdapiScreen').CAdapiScreen;
var CAdapiListOfEntity = require('syracuse-x3/lib/fusion/CAdapiListOfEntity').CAdapiListOfEntity;
var util = require('util');

var CAdapiScreenDummy = exports.CAdapiScreenDummy = helpers.defineClass(function(aRequesterWindow, aEntityIndexbaseOne, aScreenId) {
	CAdapiScreen.call(this, aRequesterWindow, aEntityIndexbaseOne, aScreenId, null);
}, CAdapiScreen, {

	traverseChilds: function(aTraverser) {
		return this.traverseChilds(aTraverser);
	},
	traverseSettable: function(aStructIst, aSetter) {
		return aSetter.setModifiable(aStructIst, this);
	}
});

var CAdapiListOfScreens = exports.CAdapiListOfScreens = helpers.defineClass(function(aRequesterWindow) {
	CAdapiListOfEntity.call(this, aRequesterWindow);
	this.screenLeftList = null;
	this.screenButtons = new CAdapiScreenDummy(aRequesterWindow, CAdapiScreen.SCREEN_BUTTONS_NUM_ZERO_BASED + 1, "_BUTTONS");
	this.screenTabTitles = new CAdapiScreenDummy(aRequesterWindow, CAdapiScreen.SCREEN_TABTITLES_NUM_ZERO_BASED + 1, "_TABTITLE");
}, CAdapiListOfEntity, {
	get: function(aIdx) {
		if (aIdx < CAdapiScreen.SCREEN_BUTTONS_NUM_ZERO_BASED && this.length > aIdx) {
			return CAdapiListOfEntity.prototype.get.call(this, aIdx);
		} else if (aIdx == CAdapiScreen.SCREEN_BUTTONS_NUM_ZERO_BASED) //32 => index base "one" = 33 => AG
		{
			return this.screenButtons;
		} else if (aIdx == CAdapiScreen.SCREEN_LEFTLIST_NUM_ZERO_BASED) //33 => index base "one" = 34 => AH
		{
			return this.screenLeftList;
		} else if (aIdx == CAdapiScreen.SCREEN_TABTITLES_NUM_ZERO_BASED) //34 => index base "one" = 35 => AI
		{
			return this.screenTabTitles;
		} else {
			return null;
		}
	},
	pushScreenLeftList: function(aScreen) {
		this.screenLeftList = aScreen;
		this.pushEntity(aScreen.id, aScreen);
	},
	traverseAll: function(aTraverser) {
		var wMax = this.length;
		var wScreen;
		var wI = 0;
		while (wI < wMax) {
			wScreen = this.get(wI);
			if (aTraverser.doWhenTraverse(wScreen, wI) == true) {
				return true;
			}
			if (wScreen.traverseAll(aTraverser) == true) {
				return true;
			}
			wI++;
		}
		return false;
	},
	traverseChilds: function(aTraverser) {
		var wMax = this.length;
		var wEntry;
		var wI = 0;
		while (wI < wMax) {
			wEntry = this.get(wI);
			if (aTraverser.doWhenTraverse(wEntry, wI) == true) {
				return true;
			}
			wI++;
		}
		return false;
	},
	traverseSettable: function(aStructIst, aSetter) {
		var wI = aStructIst.screenZeroBase;
		var wAdapiScreen = this.get(wI);
		if (wAdapiScreen == null) {
			var wMess = this.getClient().getMessage('ERROR_SCRN_UNKNOWN', CAdapiScreen.screenZeroBaseIdxToWhatScreen(wI), this.length.toString(), this.requesterWindow.id);
			aSetter.setLogicalErrMess(wMess);
			return false;
		}
		if (!aStructIst.isScreenIst()) {
			return wAdapiScreen.traverseSettable(aStructIst, aSetter);
		} else {
			return aSetter.setModifiable(aStructIst, wAdapiScreen);
		}
	}
});
