// syracuse-x3/lib/fusion/CAdapiException.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');


var CAdapiException = exports.CAdapiException = helpers.defineClass(function(obs, e, s) {
	Error.call(this, s + (e ? " [" + e.message + "]" : ""));
	this.formatException = false;
	this.inner = e;
	this.diagnose = null;
	this.setDiagnosis(obs, e, s);
	if (e) console.log("Error: " +e);
}, Error, {
	isFormatException: function() {
		return this.formatException;
	},
	setDiagnose: function(diagnose) {
		this.diagnose = diagnose;
	},
	setDiagnosis: function(obs, e, s) {
		this.diagnose = {
			$applicationCode: "ERROR",
			$code: "ApplicationDiagnosis",
			$severity: "ERROR",
			$stack: e && e.stack
		}
		/*
		if (obs != null && obs instanceof Observer) {
			this.diagnose.addObserver(obs);
			this.diagnose.forceChange();
		}*/
	},/*
	stack: {
		get: function() {
			console.log("this: " + this)
			return this.inner && this.inner.stack;
		}
	},*/

});

CAdapiException.newFmtException = function(obs, e, s) {
	var r = new CAdapiException(obs, e, s);
	r.formatException = true;
	return r;

}
