// syracuse-x3/lib/fusion/CAdapiResources.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var locale = require("syracuse-core/lib/locale");
var util = require('util');

var CAdapiResourcesHelper = exports.CAdapiResourcesHelper = helpers.defineClass(function() {

}, null, {
	getMessage: function() {
		var args = [module, "message."+arguments[0]];
		for (var i = 1 ; i < arguments.length ; i++)
			args.push(arguments[i].toString());
		var res = locale.format.apply(null, args);
		return res;
	},
});

var CAdapiResources = exports.CAdapiResources = {
	KEYMESS_ERROR_AFFSTATE: "error.affstate",
	KEYMESS_ERROR_ARG_RESU_TBL: "error.arg.resu_tbl",
	KEYMESS_ERROR_CANT_GET_PICKED_LINE_KEY: "error.cant.find.pickedline",
	KEYMESS_ERROR_CANT_GET_VAR_BLOB: "error.cant.find.varaibleblob",
	KEYMESS_ERROR_CONTEXT_NOT_VALID: "context.not.valid",
	KEYMESS_ERROR_CURFORMAT_NOFORMAT: "error.currentformat.noformat",
	KEYMESS_ERROR_CURFORMAT_NOIST: "error.currentformat.noist",
	KEYMESS_ERROR_DELETER: "error.deleter",
	KEYMESS_ERROR_DIM_OVERFLOW: "error.dim.overflow",
	KEYMESS_ERROR_DSRC_UNKNOWN: "error.dsrc.unknown",
	KEYMESS_ERROR_FORMAT_EXCEPTION: "error.format.exception",
	KEYMESS_ERROR_FORMAT_NOFORMAT: "error.format.noformat",
	KEYMESS_ERROR_INSERTER: "error.inserter",
	KEYMESS_ERROR_LINE_UNKNOWN: "error.line.unknown",
	KEYMESS_ERROR_LLIST_UNKNOWN: "error.llist.unknown",
	KEYMESS_ERROR_LLLINE_SELECTER: "error.llline.selecter",
	KEYMESS_ERROR_LOGICAL: "error.logical",
	KEYMESS_ERROR_NO_COLUMS: "no.column.in.leflist",
	KEYMESS_ERROR_NO_ISTDATA: "error.no.istdatas",
	KEYMESS_ERROR_NOT_RIGHT_MODIFIABLE: "error.not.right.modifiable",
	KEYMESS_ERROR_PIVOTDATE: "bad.pivot.date",
	KEYMESS_ERROR_RESU_NOT_NUMERIC: "error.format.resu.not.numeric",
	KEYMESS_ERROR_SCRN_UNKNOWN: "error.scrn.unknown",
	KEYMESS_ERROR_SETTER_BLOB: "error.setter.blob",
	KEYMESS_ERROR_SETTER_FORMAT: "error.setter.format",
	KEYMESS_ERROR_SETTER_STATE: "error.setter.state",
	KEYMESS_ERROR_SETTER_STYLE: "error.setter.style",
	KEYMESS_ERROR_SETTER_VALUE: "error.setter.value",
	KEYMESS_ERROR_TRAVERSE_SETTABLE: "error.traverse.settable",
	KEYMESS_ERROR_TRAVERSE_SETTABLE_CASE: "error.traverse.settable.case",
	KEYMESS_ERROR_TRAVERSE_SETTABLE_EXCEPT: "error.traverse.settable.exception",
	KEYMESS_ERROR_UNFORMAT_EXCEPTION: "error.unformat.exception",
	KEYMESS_ERROR_UPLOAD_BLOB: "error.upload.blob",
	KEYMESS_ERROR_VAR_UNKNOWN: "error.var.unknown",
	KEYMESS_ERROR_WIN_UNKNOWN: "error.ist.winAlphaId.unknown",
	KEYMESS_FORMAT_ETOS: "error.format.edit.to.show",
	KEYMESS_FORMAT_RTOS: "error.format.rought.to.show",
	KEYMESS_INFOS_ACTIVE_REQUEST: "infos.active.request",
	KEYMESS_INFOS_NOACTIVE_REQUEST: "infos.noactive.request",
	KEYMESS_LOADING_WINDOW_MESSAGE: "loading.window.mess",
	KEYMESS_LOADING_WINDOW_TITLE: "loading.window.title",
	KEYMESS_MULTIPLE_MODIFS_COLOR: "multiple.modifs.color",
	KEYMESS_MULTIPLE_MODIFS_FORMAT: "multiple.modifs.format",
	KEYMESS_MULTIPLE_MODIFS_STATE: "multiple.modifs.state",
	KEYMESS_MULTIPLE_MODIFS_STYLE: "multiple.modifs.style",
	KEYMESS_MULTIPLE_MODIFS_TITLE: "multiple.modifs.title",
	KEYMESS_MULTIPLE_MODIFS_VALUE: "multiple.modifs.value",
	KEYMESS_NDOCAR_WHAT: "ndocar.what",
	KEYMESS_SESSION_ALREADY_CLOSED: "session.already.closed",
	KEYMESS_SESSION_CANT_OPEN: "session.cant.open",
	KEYMESS_SESSION_CLOSE_MESSAGE: "session.hangup.mess",
	KEYMESS_SESSION_CLOSE_TITLE: "session.hangup.title",
	KEYMESS_UNFORMAT_STOE: "error.unformat..show.to.edit",
	KEYMESS_UNFORMAT_STOR: "error.unformat.show.to.rought",
	RES_ID: "com.sage.x3.bundle.shared.adapi.CAdapi",
};


function CAdapiResources(aLang) {
	//TODO_SUPER(RES_ID, aLang);
}
