// syracuse-x3/lib/fusion/CAdapiListOfLocalMenusFactory.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Hashtable = Object; //require('syracuse-x3/lib/fusion/Hashtable').Hashtable;

var CAdapiListOfLocalMenus = exports.CAdapiListOfLocalMenus = helpers.defineClass(function() {
	Hashtable.call(this);
}, Hashtable, {
	addLocalMenu: function(aAdapiLocalMenu) {
		this.put(aAdapiLocalMenu.numId, aAdapiLocalMenu);
	}
});

var MESS = "unknown choice=[%s] : menu [%s] has [%s] items";
var CAdapiLocalMenu = exports.CAdapiLocalMenu = helpers.defineClass(function(aNumId, aId, aChoices) {
	this.numId = aNumId;
	this.id = aId;
	this.choices = aChoices;
}, null, {
	getLib: function(aChoice) {
		if (aChoice < 1 || aChoice > this.choices.length) {
			var wMess = CStringFacilities.sprintf(MESS, aChoice.toString(), this.id, this.choices.length.toString());
			throw new Exception(wMess);
		}
		return this.choices[aChoice - 1]; // base 0 !
	},
});

var ATT_NAM = "NAM";
var ATT_NBZ = "NBZ";
var TAG_MENL = "MENL";
var TAG_V = "V";

function extractIntLocalMenuId(aLocalMenuId) {
	aLocalMenuId = aLocalMenuId.substring(1); //pour enlever "M"
	return parseInt(aLocalMenuId);
}

function instanciateLocalMenu(aAdapiClient, aId, aLocalMenuID, aNbChoices, aLocalMenuNode) {
	var wLocalMenu;
	try {
		var wChoicesNodes = aLocalMenuNode[TAG_V];
		var wChoiceNode;
		var wMax = Array.isArray(wChoicesNodes) ? wChoicesNodes.length : 1;
		var wChoices = [];
		var wI = 0;
		while (wI < wMax) {
			wChoiceNode = wMax === 1 ? wChoicesNodes : wChoicesNodes[wI];
			wChoices[wI] = wChoiceNode;
			wI++;
		}
		wLocalMenu = new CAdapiLocalMenu(aId, aLocalMenuID, wChoices);
	} catch (e) {
		throw new CAdapiException(aAdapiClient, e, "CANT_ADD_LOCALMENU - instanciateLocalMenu");
	}
	return wLocalMenu;
}

var CAdapiListOfLocalMenusFactory = {

	instanciateLocalMenus: function(aAdapiRequesterWindow, aWindowDescription, aLocalMenus) {
		var wLocalMenusNodes = aWindowDescription.MENLOC[TAG_MENL];
		var wMax = Array.isArray(wLocalMenusNodes) ? wLocalMenusNodes.length : 1;
		var wI = 0;
		var wLocalMenuId;
		var wLocalMenuNode;
		var wNbChoices;
		while (wI < wMax) {
			wLocalMenuNode = wMax === 1 ? wLocalMenusNodes : wLocalMenusNodes[wI];
			wLocalMenuId = wLocalMenuNode.$[ATT_NAM];
			wNbChoices = wLocalMenuNode.$[ATT_NBZ];
			var wNumId = extractIntLocalMenuId(wLocalMenuId);
			if (aLocalMenus.get(wNumId) != null) {
				throw new CAdapiException(aAdapiRequesterWindow.adapiClient, "CANT_ADD_LOCALMENU - AlreadyExist");
			}
			aLocalMenus.addLocalMenu(instanciateLocalMenu(aAdapiRequesterWindow.adapiClient, wNumId, wLocalMenuId, wNbChoices, wLocalMenuNode));
			wI++;
		}
	}
}
