// syracuse-x3/lib/fusion/CAdapiLeftList.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var CAdapiDataSrc = require('syracuse-x3/lib/fusion/CAdapiDataSrc').CAdapiDataSrc;

//private static int FMT_FLAT = 1;
var FMT_HIERARCHIC = 2;
var FMT_PICKING_FLAT = 5;
var FMT_PICKING_HIERARCHIC = 3;
var FMT_RECURSIVE = 4;

var CAdapiLeftList = exports.CAdapiLeftList = helpers.defineClass(function(aStackedWindow, aScreen, aEntityIndexBaseOne, aId, aListFormat, aNbColumns) {
	CAdapiDataSrc.call(this, aStackedWindow, aScreen, aEntityIndexBaseOne, aId, 300, aNbColumns, null, CAdapiDataSrc.TYPE_LEFTLIST);
	this.listFormat = aListFormat;
	this.columnTitleModified = false;
	this.currentSelectedLine = -1;
	this.nbColumnModified = false;
	this.colomnFormatChange = false;
}, CAdapiDataSrc, {
	appendLeftListLine: function(aIdx, aBranche, aTxt, aVal, aClf, aClp) {
		var wLeftListLine = this.entrySup.clone();
		if (aIdx == -1) {
			aIdx = this.lines.length;
		}
		wLeftListLine.setColumns(aIdx - 1, aVal, aClf);
		CAdapiDataSrc.prototype.appendLeftListLine.call(this, wLeftListLine);
		return wLeftListLine;
	},
	changeColumnsFormats: function(aFormats) {
		var CAdapiListOfLeftListsFactory = require('syracuse-x3/lib/fusion/CAdapiListOfLeftListsFactory').CAdapiListOfLeftListsFactory;
		var wLeftListLine;
		if (aFormats.length > this.nbColumns) {
			wLeftListLine = CAdapiListOfLeftListsFactory.instanciateNewLeftListEntrySup(this.stackedWindow, this, aFormats.length);
			this.setEntrySup(wLeftListLine);
			this.nbColumnModified = true;
		} else {
			wLeftListLine = this.entrySup;
		}
		wLeftListLine.setColumnsFormats(aFormats);
		this.colomnFormatChange = true;
	},
	changeColumnsTitles: function(aTitles) {
		this.entrySup.setColumnsTitles(aTitles);
		this.columnTitleModified = true;
	},
	findLineIdx: function(aCle) {
		var wCle;
		var wMax = this.lines.length;
		var wI = 0;
		while (wI < wMax) {
			wCle = this.getLine(wI).cle;
			if (wCle == null) {

			} else if (wCle === aCle) {
				return wI;
			}
			wI++;
		}
		return -999;
	},
	hasColumnTitleModified: function() {
		return this.columnTitleModified;
	},
	hasNbColumnModified: function() {
		return this.nbColumnModified;
	},
	isColomnFormatChange: function() {
		return this.colomnFormatChange;
	},
	isHierarchic: function() {
		return CAdapiLeftList.isHierarchicFormat(this.listFormat);
	},
	isListFormat: function(aListFormat) {
		return this.listFormat == aListFormat;
	},
	isRecursive: function() {
		return CAdapiLeftList.isRecursiveFormat(this.listFormat);
	},
	razModification: function() {
		this.nbColumnModified = false;
		this.columnTitleModified = false;

		CAdapiDataSrc.prototype.razModification.call(this);
	},
	setColomnFormatChange: function(colomnFormatChange) {
		this.colomnFormatChange = colomnFormatChange;
	},
	setServerSelectLine: function(selected, aCle) {
		var selIdx = this.findLineIdx(aCle);
		var found = selIdx > -1;
		if (this.isListFormat(FMT_PICKING_FLAT)) {
			if (found) {
				this.getLine(selIdx).setLeftListLineSelected(selected, true);
			}
		} else {
			if (selected && selIdx != this.currentSelectedLine) {
				if (selIdx > -1) {
					this.getLine(selIdx).setLeftListLineSelected(selected, true);
					if (this.currentSelectedLine > -1 && this.currentSelectedLine < this.lines.length) {
						this.getLine(this.currentSelectedLine).setLeftListLineSelected(false, false);
					}
				} else if (this.currentSelectedLine > -1 && this.currentSelectedLine < this.lines.length) {
					this.getLine(this.currentSelectedLine).setLeftListLineSelected(false, true);

				}
				this.currentSelectedLine = selIdx;

			} else {
				if (selIdx > -1) {
					this.getLine(selIdx).setLeftListLineSelected(selected, true);
				}
				this.currentSelectedLine = -1;
			}
		}
		return true;
	}

});

CAdapiLeftList.isHierarchicFormat = function(aFormat) {
	return aFormat == FMT_HIERARCHIC || aFormat == FMT_PICKING_HIERARCHIC;
}

CAdapiLeftList.isRecursiveFormat = function(aFormat) {
	return aFormat == FMT_RECURSIVE;
}
