// syracuse-ui/lib/tests/pages/extend/landingCalendar.js
"use strict";

exports.load = function(page, finish) {
	page.setTitle("Landing Page Calendar");
	page.$prototype.$properties = {
		agenda: {
			$title: "Agenda",
			$type: "application/x-array",
			$item: {
				$properties: {
					$criteria: {
						$type: "application/x-filter",
						$item: {
							$properties: {
								"DATSTR": {},
								"DATEND": {},
								"displayLineNum": {},
								"recordsPerPage": {},
								"req_id": {},
								"ifCube": {}
							}
						}
					},
					"line_no": {
						$title: "Line no.",
						$type: "application/x-integer"
					},
					"z_p": {
						$type: "application/x-integer",
						"$isExcluded": true
					},
					"z_m": {
						$type: "application/x-integer",
						"$isExcluded": true
					},
					"CLE": {
						$title: "Clé",
						$type: "application/x-string",
						"$maxLength": 10
					},
					"TEXTE": {
						$title: "Normal text",
						$type: "application/x-string",
						"$maxLength": 20
					},
					"FCY": {
						$title: "Site",
						$type: "application/x-string",
						"$maxLength": 5,
						"$links": {
							"$details": {
								$title: "Detail",
								$type: "application/json;vnd.sage=syracuse",
								"$url": "{$baseUrl}/FACILITY('{FCY}')?representation=FACILITY.$details",
								"$target": "main"
							}
						}
					},
					"FDAT": {
						$title: "Start date",
						$type: "application/x-date"
					},
					"LDAT": {
						$title: "End date",
						$type: "application/x-date"
					},
					"HOUR1": {
						$title: "Hour minute",
						$type: "application/x-string",
						$maxLength: 5
					},
					"HOUR2": {
						$title: "Hours (worked) other",
						$type: "application/x-string",
						$maxLength: 5
					},
					"NUMERIC": {
						$title: "Decimal",
						$type: "application/x-decimal",
						"$format": "#,###.00",
						"$precision": 11,
						"$scale": 2
					},
					"TEXTE1": {
						$title: "Additional text",
						$type: "application/x-string",
						$maxLength: 15
					},
					"CREDATTIM": {
						$title: "Date time",
						$type: "application/x-datetime"
					},
					"LONGLONGFIEL": {
						$title: "Long",
						$type: "application/x-string",
						$maxLength: 10
					},
					"MENLOC": {
						$title: "Menu",
						$type: "application/x-choice",
						"$value": {
							$type: "application/x-integer",
							"$enum": [{
								"$value": 1,
								$title: "Non-existent Object"
							}, {
								"$value": 2,
								$title: "Non-existent Language"
							}, {
								"$value": 3,
								$title: "Warning, this code does not match any X3 user code"
							}]
						}
					},
					"AUUID": {
						$title: "Single identifier",
						$type: "application/x-string",
						$maxLength: 36
					}
				}
			}
		}
	};
	syra_widgets.ensureMediaTypes(page.$prototype, function() {
		page.addItem(page.layoutContent.domItem, {
			"$bind": "agenda"
		});
		var _start = "2015-06";
		var _end = "2015-07";
		page.applyChange({
			agenda: {
				"$calendars": [{
					"description": "default calendar",
					"id": "1",
					"$category": 0,
					"$pageFilter": {
						"start": ["DATSTR", "HOUSTR"],
						"end": ["DATSTR", "HOUEND"]
					},
					"$categoryLegend": {
						"property": "category",
						"legend": {
							"1": "Normal event",
							"2": "Out of office",
							"3": "Busy",
							"4": "Do not disturb",
							"5": "Sick leave",
							"6": "In prison",
							"7": "Abdicated",
							"8": "Normal event",
							"9": "Out of office",
							"10": "Busy again",
							"11": "Really annoyed",
							"12": "Smashed",
							"13": "Shattered",
							"14": "Hammered",
							"15": "En panne",
							"16": "Péter un cable"
						}
					},
					"$colorLegend": {
						"property": "color",
						"legend": {
							"1": "Application",
							"2": "Development",
							"3": "Source-control",
							"4": "System",
							"5": "Administration",
							"6": "Testing",
							"7": "Verification",
							"8": "Industrialisation",
							"9": "Sales",
							"10": "Inventory",
							"11": "Returns",
							"12": "Nominal-ledger",
							"13": "Sales-ledger",
							"14": "Purchase-ledger",
							"15": "Printing",
							"16": "Storage"
						}
					}
				}],
				"$events": [{
					"$category": 0,
					"$calendar": 0,
					"id": "77851f9d-bd26-4dc1-a6c2-b4f4b98006ab",
					"summary": "Random text BUTRE",
					"subtitle": "A temporary Subtitle!",
					"description": {
						"CLE": "CLE000001",
						"TEXTE1": "Rnd # 64742",
						"FCY": "ZUSA"
					},
					"start": {
						"datetime": _start + "-01T08:00Z"
					},
					"end": {
						"datetime": _start + "-01T09:00Z"
					},
					"color": 1,
					"category": 1
				}, {
					"$category": 0,
					"$calendar": 0,
					"id": "f414313f-b352-466c-af45-7e9f107157cd",
					"summary": "Random text BBDIU",
					"subtitle": "A temporary Subtitle!",
					"description": {
						"CLE": "CLE000002",
						"TEXTE1": "Rnd # 35500",
						"FCY": "SCI"
					},
					"start": {
						"datetime": _start + "-02T08:00Z"
					},
					"end": {
						"datetime": _start + "-02T23:59Z"
					},
					"color": 2,
					"category": 2
				}, {
					"$category": 0,
					"$calendar": 0,
					"id": "f414313f-b352-466c-af45-7e9f107157cd",
					"summary": "Random text BBDIU",
					"subtitle": "A temporary Subtitle!",
					"description": {
						"CLE": "CLE000002",
						"TEXTE1": "Rnd # 35501",
						"FCY": "SCI"
					},
					"start": {
						"datetime": _start + "-03T08:00Z"
					},
					"end": {
						"datetime": _start + "-03T23:59Z"
					},
					"color": 3,
					"category": 3
				}, {
					"$category": 0,
					"$calendar": 0,
					"id": "f414313f-b352-466c-af45-7e9f107157cd",
					"summary": "Random text BBDIU",
					"subtitle": "A temporary Subtitle!",
					"description": {
						"CLE": "CLE000002",
						"TEXTE1": "Rnd # 35503",
						"FCY": "SCI"
					},
					"start": {
						"datetime": _start + "-03T08:05Z"
					},
					"end": {
						"datetime": _start + "-03T23:59Z"
					},
					"color": 2,
					"category": 2
				}, {
					"$category": 0,
					"$calendar": 0,
					"id": "f414313f-b352-466c-af45-7e9f107157cd",
					"summary": "Random text BBDIU",
					"subtitle": "A temporary Subtitle!",
					"description": {
						"CLE": "CLE000002",
						"TEXTE1": "Rnd # 35504",
						"FCY": "SCI"
					},
					"start": {
						"datetime": _start + "-03T08:10Z"
					},
					"end": {
						"datetime": _start + "-03T23:59Z"
					},
					"color": 1,
					"category": 1
				}, {
					"$category": 0,
					"$calendar": 0,
					"id": "f414313f-b352-466c-af45-7e9f107157cd",
					"summary": "Random text BBDIU",
					"subtitle": "A temporary Subtitle!",
					"description": {
						"CLE": "CLE000002",
						"TEXTE1": "Rnd # 35505",
						"FCY": "SCI"
					},
					"start": {
						"datetime": _start + "-03T08:15Z"
					},
					"end": {
						"datetime": _start + "-03T23:59Z"
					},
					"color": 4,
					"category": 4
				}, {
					"$category": 0,
					"$calendar": 0,
					"id": "f414313f-b352-466c-af45-7e9f107157cd",
					"summary": "Random text BBDIU",
					"subtitle": "A temporary Subtitle!",
					"description": {
						"CLE": "CLE000002",
						"TEXTE1": "Rnd # 35506",
						"FCY": "SCI"
					},
					"start": {
						"datetime": _start + "-03T08:20Z"
					},
					"end": {
						"datetime": _start + "-03T23:59Z"
					},
					"color": 5,
					"category": 5
				}, {
					"$category": 0,
					"$calendar": 0,
					"id": "f414313f-b352-466c-af45-7e9f107157cd",
					"summary": "Random text BBDIU",
					"subtitle": "A temporary Subtitle!",
					"description": {
						"CLE": "CLE000002",
						"TEXTE1": "Rnd # 35507",
						"FCY": "SCI"
					},
					"start": {
						"datetime": _start + "-03T08:25Z"
					},
					"end": {
						"datetime": _start + "-03T23:59Z"
					},
					"color": 6,
					"category": 6
				}, {
					"$category": 0,
					"$calendar": 0,
					"id": "f414313f-b352-466c-af45-7e9f107157cd",
					"summary": "Random text BBDIU",
					"subtitle": "A temporary Subtitle!",
					"description": {
						"CLE": "CLE000002",
						"TEXTE1": "Rnd # 35508",
						"FCY": "SCI"
					},
					"start": {
						"datetime": _start + "-03T09:30Z"
					},
					"end": {
						"datetime": _start + "-03T23:59Z"
					},
					"color": 7,
					"category": 7
				}, {
					"$category": 0,
					"$calendar": 0,
					"id": "f414313f-b352-466c-af45-7e9f107157cd",
					"summary": "Random text BBDIU",
					"subtitle": "A temporary Subtitle!",
					"description": {
						"CLE": "CLE000002",
						"TEXTE1": "Rnd # 35509",
						"FCY": "SCI"
					},
					"start": {
						"datetime": _start + "-03T09:35Z"
					},
					"end": {
						"datetime": _start + "-03T23:59Z"
					},
					"color": 8,
					"category": 8
				}, {
					"$category": 0,
					"$calendar": 0,
					"id": "f414313f-b352-466c-af45-7e9f107157cd",
					"summary": "Random text BBDIU",
					"subtitle": "A temporary Subtitle!",
					"description": {
						"CLE": "CLE000002",
						"TEXTE1": "Rnd # 35510",
						"FCY": "SCI"
					},
					"start": {
						"datetime": _start + "-03T09:40Z"
					},
					"end": {
						"datetime": _start + "-03T23:59Z"
					},
					"color": 7,
					"category": 7
				}, {
					"$category": 0,
					"$calendar": 0,
					"id": "f414313f-b352-466c-af45-7e9f107157cd",
					"summary": "Random text BBDIU",
					"subtitle": "A temporary Subtitle!",
					"description": {
						"CLE": "CLE000002",
						"TEXTE1": "Rnd # 35511",
						"FCY": "SCI"
					},
					"start": {
						"datetime": _start + "-03T09:45Z"
					},
					"end": {
						"datetime": _start + "-03T23:59Z"
					},
					"color": 6,
					"category": 6
				}, {
					"$category": 0,
					"$calendar": 0,
					"id": "f414313f-b352-466c-af45-7e9f107157cd",
					"summary": "Random text BBDIU",
					"subtitle": "A temporary Subtitle!",
					"description": {
						"CLE": "CLE000002",
						"TEXTE1": "Rnd # 35512",
						"FCY": "SCI"
					},
					"start": {
						"datetime": _start + "-03T09:50Z"
					},
					"end": {
						"datetime": _start + "-03T23:59Z"
					},
					"color": 5,
					"category": 5
				}, {
					"$category": 0,
					"$calendar": 0,
					"id": "f414313f-b352-466c-af45-7e9f107157cd",
					"summary": "Random text BBDIU",
					"subtitle": "A temporary Subtitle!",
					"description": {
						"CLE": "CLE000002",
						"TEXTE1": "Rnd # 35512",
						"FCY": "SCI"
					},
					"start": {
						"datetime": _start + "-03T09:55Z"
					},
					"end": {
						"datetime": _start + "-03T23:59Z"
					},
					"color": 6,
					"category": 6
				}, {
					"$category": 0,
					"$calendar": 0,
					"id": "e8bbafd0-cc68-48f0-a728-f4853a700bb7",
					"summary": "Random text LNSOG",
					"subtitle": "A temporary Subtitle!",
					"description": {
						"CLE": "CLE000010",
						"TEXTE1": "Rnd # 85472",
						"FCY": "ZUS"
					},
					"start": {
						"datetime": _start + "-04T06:00Z"
					},
					"end": {
						"datetime": _start + "-04T07:52Z"
					},
					"color": 4,
					"category": 4
				}, {
					"$category": 0,
					"$calendar": 0,
					"id": "e8bbafd0-cc68-48f0-a728-f4853a700bb7",
					"summary": "Random text LNSOG",
					"subtitle": "A temporary Subtitle!",
					"description": {
						"CLE": "CLE000010",
						"TEXTE1": "Rnd # 85473",
						"FCY": "ZUS"
					},
					"start": {
						"datetime": _start + "-05T06:00Z"
					},
					"end": {
						"datetime": _start + "-05T07:52Z"
					},
					"color": 5,
					"category": 5
				}, {
					"$category": 0,
					"$calendar": 0,
					"id": "e8bbafd0-cc68-48f0-a728-f4853a700bb7",
					"summary": "Random text LNSOG",
					"subtitle": "A temporary Subtitle!",
					"description": {
						"CLE": "CLE000010",
						"TEXTE1": "Rnd # 85474",
						"FCY": "ZUS"
					},
					"start": {
						"datetime": _start + "-06T06:00Z"
					},
					"end": {
						"datetime": _start + "-06T07:52Z"
					},
					"color": 6,
					"category": 6
				}, {
					"$category": 0,
					"$calendar": 0,
					"id": "e8bbafd0-cc68-48f0-a728-f4853a700bb7",
					"summary": "Random text LNSOG",
					"subtitle": "A temporary Subtitle!",
					"description": {
						"CLE": "CLE000010",
						"TEXTE1": "Rnd # 85475",
						"FCY": "ZUS"
					},
					"start": {
						"datetime": _start + "-07T06:00Z"
					},
					"end": {
						"datetime": _start + "-07T07:52Z"
					},
					"color": 7,
					"category": 7
				}, {
					"$category": 0,
					"$calendar": 0,
					"id": "e8bbafd0-cc68-48f0-a728-f4853a700bb7",
					"summary": "Random text LNSOG",
					"subtitle": "A temporary Subtitle!",
					"description": {
						"CLE": "CLE000010",
						"TEXTE1": "Rnd # 85476",
						"FCY": "ZUS"
					},
					"start": {
						"datetime": _start + "-08T06:00Z"
					},
					"end": {
						"datetime": _start + "-08T07:52Z"
					},
					"color": 8,
					"category": 8
				}, {
					"$category": 0,
					"$calendar": 0,
					"id": "e8bbafd0-cc68-48f0-a728-f4853a700bb7",
					"summary": "Random text LNSOG",
					"subtitle": "A temporary Subtitle!",
					"description": {
						"CLE": "CLE000010",
						"TEXTE1": "Rnd # 85478",
						"FCY": "ZUS"
					},
					"start": {
						"datetime": _start + "-09T06:00Z"
					},
					"end": {
						"datetime": _start + "-09T07:52Z"
					},
					"color": 9,
					"category": 9
				}, {
					"$category": 0,
					"$calendar": 0,
					"id": "e8bbafd0-cc68-48f0-a728-f4853a700bb7",
					"summary": "Random text LNSOG",
					"subtitle": "A temporary Subtitle!",
					"description": {
						"CLE": "CLE000010",
						"TEXTE1": "Rnd # 85479",
						"FCY": "ZUS"
					},
					"start": {
						"datetime": _start + "-10T06:00Z"
					},
					"end": {
						"datetime": _start + "-10T07:52Z"
					},
					"color": 10,
					"category": 10
				}, {
					"$category": 0,
					"$calendar": 0,
					"id": "e8bbafd0-cc68-48f0-a728-f4853a700bb7",
					"summary": "Random text LNSOG",
					"subtitle": "A temporary Subtitle!",
					"description": {
						"CLE": "CLE000010",
						"TEXTE1": "Rnd # 85480",
						"FCY": "ZUS"
					},
					"start": {
						"datetime": _start + "-11T06:00Z"
					},
					"end": {
						"datetime": _start + "-11T07:52Z"
					},
					"color": 11,
					"category": 11
				}, {
					"$category": 0,
					"$calendar": 0,
					"id": "e8bbafd0-cc68-48f0-a728-f4853a700bb7",
					"summary": "Random text LNSOG",
					"subtitle": "A temporary Subtitle!",
					"description": {
						"CLE": "CLE000010",
						"TEXTE1": "Rnd # 85481",
						"FCY": "ZUS"
					},
					"start": {
						"datetime": _start + "-12T06:00Z"
					},
					"end": {
						"datetime": _start + "-12T07:52Z"
					},
					"color": 12,
					"category": 12
				}, {
					"$category": 0,
					"$calendar": 0,
					"id": "e8bbafd0-cc68-48f0-a728-f4853a700bb7",
					"summary": "Random text LNSOG",
					"subtitle": "A temporary Subtitle!",
					"description": {
						"CLE": "CLE000010",
						"TEXTE1": "Rnd # 85482",
						"FCY": "ZUS"
					},
					"start": {
						"datetime": _start + "-13T06:00Z"
					},
					"end": {
						"datetime": _start + "-13T07:52Z"
					},
					"color": 13,
					"category": 13
				}, {
					"$category": 0,
					"$calendar": 0,
					"id": "e8bbafd0-cc68-48f0-a728-f4853a700bb7",
					"summary": "Random text LNSOG",
					"subtitle": "A temporary Subtitle!",
					"description": {
						"CLE": "CLE000010",
						"TEXTE1": "Rnd # 85483",
						"FCY": "ZUS"
					},
					"start": {
						"datetime": _start + "-14T06:00Z"
					},
					"end": {
						"datetime": _start + "-14T07:52Z"
					},
					"color": 14,
					"category": 14
				}, {
					"$category": 0,
					"$calendar": 0,
					"id": "e8bbafd0-cc68-48f0-a728-f4853a700bb7",
					"summary": "Random text LNSOG",
					"subtitle": "A temporary Subtitle!",
					"description": {
						"CLE": "CLE000010",
						"TEXTE1": "Rnd # 85484",
						"FCY": "ZUS"
					},
					"start": {
						"datetime": _start + "-15T06:00Z"
					},
					"end": {
						"datetime": _start + "-15T07:52Z"
					},
					"color": 15,
					"category": 15
				}, {
					"$category": 0,
					"$calendar": 0,
					"id": "e8bbafd0-cc68-48f0-a728-f4853a700bb7",
					"summary": "Random text LNSOG",
					"subtitle": "A temporary Subtitle!",
					"description": {
						"CLE": "CLE000010",
						"TEXTE1": "Rnd # 85485",
						"FCY": "ZUS"
					},
					"start": {
						"datetime": _start + "-16T06:00Z"
					},
					"end": {
						"datetime": _start + "-16T07:52Z"
					},
					"color": 16,
					"category": 16
				}, {
					"$category": 0,
					"$calendar": 0,
					"id": "e8bbafd0-cc68-48f0-a728-f4853a700bb7",
					"summary": "A Very long event",
					"subtitle": "A temporary Subtitle!",
					"description": {
						"CLE": "CLE000010",
						"TEXTE1": "Rnd # 85486",
						"FCY": "ZUS"
					},
					"start": {
						"datetime": _start + "-16T06:00Z"
					},
					"end": {
						"datetime": _end + "-16T07:52Z"
					},
					"color": 2,
					"category": 2
				}]
			}
		});
		finish(page);
	});
};
