// syracuse-ui/lib/tests/pages/array/graph.js
"use strict";
exports.load = function(page, finish) {
	page.refreshData = function() {
		var records = [{
			$uuid: "d3e97372-21c1-40ac-b147-4a0eaad4f58e",
			$key: "d3e97372-21c1-40ac-b147-4a0eaad4f58e",
			ARTICLE: "Toto",
			code: "TOOT",
			matiere_premiere: "10.3",
			main_oeuvre: "20.5",
			cout_machine: "56.3",
			cout_machine2: "46.3",
			cout_machine3: "60.3"
		}, {
			$uuid: "a3e97372-21c1-40ac-b147-4a0eaad4f58e",
			$key: "a3e97372-21c1-40ac-b147-4a0eaad4f58e",
			ARTICLE: "Cle USB",
			code: "CLEUSB",
			matiere_premiere: "12.3",
			main_oeuvre: "45.2",
			cout_machine: "125.2",
			cout_machine2: "120.2",
			cout_machine3: "130.3"
		}, {
			$uuid: "b3e97372-21c1-40ad-b147-4a0eaad4f58e",
			$key: "b3e97372-21c1-40ad-b147-4a0eaad4f58e",
			ARTICLE: "Ecran Plasma",
			code: "ECRPLASMA",
			matiere_premiere: "125.2",
			main_oeuvre: "50.3",
			cout_machine: "100.1",
			cout_machine2: "90.1",
			cout_machine3: "120.3"
		}, {
			$uuid: "c3e97372-21c2-40ac-b147-4a0eaad4f58e",
			$key: "c3e97372-21c2-40ac-b147-4a0eaad4f58e",
			ARTICLE: "Lecteur DVD",
			code: "Lecteur DVD",
			matiere_premiere: "45.2",
			main_oeuvre: "23.3",
			cout_machine: "22.3",
			cout_machine2: "12.3",
			cout_machine3: "32.3"
		}, {
			$uuid: "d4e97342-21c1-40ac-b147-4a0eaad4f58e",
			$key: "d4e97342-21c1-40ac-b147-4a0eaad4f58e",
			ARTICLE: "Imprimante Laser",
			code: "IMPLASER",
			matiere_premiere: "45.2",
			main_oeuvre: "65.00",
			cout_machine: "45.25",
			cout_machine2: "40.25",
			cout_machine3: "50.25"
		}, {
			$uuid: "d5e97372-22c1-40ac-b147-4a0eaad4f58e",
			$key: "d5e97372-22c1-40ac-b147-4a0eaad4f58e",
			ARTICLE: "CD-ROM",
			code: "CDROM",
			matiere_premiere: "2.0",
			cout_machine: "100.5",
			cout_machine2: "200.3",
			cout_machine3: "120.3"
		}];
		page.applyChange({
			$properties: {
				xarray: {
					$totalResults: records.length
				}
			},
			xarray: records
		});
	};
	page.$prototype.$properties.xarray = {
		$type: "application/x-array",
		$axes: [{
			$title: "Articles vendus",
			$hierarchies: [
				["article_h"]
			]
		}, {
			$title: "measures",
			$hierarchies: [
				["$measures"]
			]
		}],
		$cube: {
			$mode: "table",
			$style: "xy",
			$displaysOneMeasure: false,
			$title: "Coûts articles",
			$hierarchies: {
				"article_h": {
					$title: "Articles",
					"$properties": ["ARTICLE"],
					"$members": [],
					"$isDefault": true
				}
			},
			$measures: {
				matiere_premiere: {
					$property: "matiere_premiere",
					$title: "Matière premiere",
					$mode: "sum",
					"$style": "line",
					"$isStacked": false,
					"$isNormalized": false,
					"$isDefault": true
				},
				main_oeuvre: {
					$property: "main_oeuvre",
					$title: "Main d'oeuvre",
					$mode: "sum",
					$style: "stick",
					$isStacked: false,
					$isNormalized: false
				},
				cout_machine: {
					$property: "cout_machine",
					$title: "Cout machine",
					$mode: "sum",
					$style: "spline",
					$isStacked: false,
					$isNormalized: false
				}
			}
		},
		$item: {
			$properties: {
				ARTICLE: {
					$title: "Intitulé article",
					$type: "application/x-string"
				},
				code: {
					$title: "Code article",
					$type: "application/x-string"
				},
				matiere_premiere: {
					$title: "Matiere première",
					$type: "application/x-decimal"
				},
				main_oeuvre: {
					$title: "Main d'oeuvre",
					$type: "application/x-decimal"
				},
				cout_machine: {
					$title: "Coût machine",
					$type: "application/x-decimal"
				},
				cout_machine2: {
					$title: "Coût machine 2",
					$type: "application/x-decimal"
				},
				cout_machine3: {
					$title: "Coût machine 3",
					$type: "application/x-decimal"
				}
			}
		}
	};
	syra_widgets.ensureMediaTypes(page.$prototype, function() {
		page.loadNewItem(page.leftCol, {
			$title: "bottom",
			$graphPosition: "bottom",
			$bind: "xarray",
			$cube: {
				$measures: {
					matiere_premiere: {
						$style: "pie"
					},
					main_oeuvre: {
						$style: "areaspline"
					},
					cout_machine: {
						$style: "bar"
					}
				}
			}
		});
		page.loadNewItem(page.rightCol, {
			$title: "bottom",
			$graphPosition: "bottom",
			$bind: "xarray",
			$cube: {
				"$style": "spiderweb",
				"$measures": {
					matiere_premiere: {
						"$property": "matiere_premiere",
						"$title": "Matière premiere",
						"$style": "spiderweb",
					},
					main_oeuvre: {
						"$property": "main_oeuvre",
						"$title": "Main d'oeuvre",
						"$isStacked": true,
						"$style": "spiderweb",
					},
					cout_machine: {
						"$property": "cout_machine",
						"$title": "Cout machine",
						"$isStacked": true,
						"$style": "spiderweb",
					},
					cout_machine2: {
						"$property": "cout_machine2",
						$title: "Cout machine 2"
					},
					cout_machine3: {
						"$property": "cout_machine3",
						$title: "Cout machine 3"
					}
				}
			}
		});
		["left", "right"].forEach(function($position) {
			page.loadNewItem(page[$position + "Col"], {
				$title: $position,
				$graphPosition: $position,
				$bind: "xarray",
				$cube: {
					$measures: {
						matiere_premiere: {
							$style: "column"
						},
						main_oeuvre: {
							$style: "line"
						},
						cout_machine: {
							$style: "scatter"
						}
					}
				}
			});
		});
		page.loadNewItem(page.leftCol, {
			$title: "behind",
			$bind: "xarray",
			$graphPosition: "behind"
		});
		page.loadNewItem(page.rightCol, {
			$title: "front",
			$bind: "xarray",
			$graphPosition: "front"
		});

		[true, false].forEach(function($isNormalized) {
			page.loadNewItem(page.leftCol, {
				$title: "bottom",
				$graphPosition: "front",
				$bind: "xarray",
				$cube: {
					"$style": "column",
					"$measures": {
						matiere_premiere: {
							"$property": "matiere_premiere",
							"$title": "Matière premiere",
							"$isStacked": true,
							$style: "column",
							"$isNormalized": $isNormalized,
							"$stackingGroup": 0
						},
						main_oeuvre: {
							"$property": "main_oeuvre",
							"$title": "Main d'oeuvre",
							"$isStacked": true,
							$style: "column",
							"$isNormalized": $isNormalized,
							"$stackingGroup": 0
						},
						cout_machine: {
							"$property": "cout_machine",
							"$title": "Cout machine",
							"$isStacked": true,
							$style: "column",
							"$isNormalized": $isNormalized,
							"$stackingGroup": 0
						}
					}
				}
			});
		});
		page.leftCol.style.width = "50%";
		page.rightCol.style.width = "50%";
		page.refreshData();
		finish(page);
	});
};
