// @sage/syracuse-lib/src/collaboration/entities/services/mongo-service.ts
const locale = require('streamline-locale');

export const entity = {
	$titleTemplate: "Mongo DB service",
	$valueTemplate: "{description}",
	$helpPage: "Administration-reference_mongo-service",
	$properties: {
		description: {
			$title: "Description",
			$isMandatory: true,
			$linksToDetails: true
		},
		url: {
			$title: "Connection URL",
            $isMandatory: true,
            $default: 'mongodb://localhost:27017'
		},
	},
	$relations: {},
	$services: {
  		test: {
			$method: "POST",
			$isMethod: true,
			$title: "Test connection",
			$execute(_, context, instance) {
				try {
					const client = require('mongodb').MongoClient.connect(instance.url(_), _);
					const result = client.admin().listDatabases(_);
					client.close(_);
					if (Array.isArray(result.databases)) instance.$addDiagnose('success', `connection test succeeded`);
					else throw new Error(`unexpected result: ${JSON.stringify(result)}`);
				} catch (ex) {
					instance.$addDiagnose("error", ex.message);
				}
			}
		},
	},
	$events: {},
	$functions: {
		getKind() { return 'mongo'; }
	},

};
