// syracuse-ui/lib/page/landing/vignette/field/boWebIVignetteField.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var VignetteField = require("syracuse-ui/lib/page/landing/vignette/field/vignetteField").VignetteField;

var TUNNEL_DESC = {
	"spgm": "INTRUTILA:BOLINK",
	"params": [{ // FONCTION
		"dim": 1,
		"typ": "STRU",
		"siz": 8
	}, { // VARIA
		"dim": 1,
		"typ": "STRU",
		"siz": 8
	}, { // CLE
		"dim": 1,
		"typ": "STRU",
		"siz": 250
	}]
};

function BOWebIVignetteField() {}

exports.BOWebIVignetteField = helpers.defineClass(BOWebIVignetteField, VignetteField, {
	ensureDefaultTitle: function() {

	},
	setDataBind: function(value, record, metaData, $bind) {
		if (this.page.externalAdapter.setDataBind(this, value, record, metaData)) {
			if (value !== undefined) {
				this.$contentUrl = this.$item.$url = value;
				this._showUrl();
			}
		}
	},
	applyDesignMetaData: function(metaData, designing) {
		VignetteField.prototype.applyDesignMetaData.call(this, metaData, designing);
		if (metaData.$url !== undefined) {
			this.$contentUrl = this.$item.$url = metaData.$url;
			this._showUrl();
		}
	},
	_showUrl: function() {
		var self = this;
		if (self.$contentUrl) {
			if (!self.boWindow) {
				window.x3boSendAckCallTunnel = function(func, value, key) {
					setTimeout(function() {
						self.externalCall(TUNNEL_DESC, [func, value, key]);
						if (self.boWindow) {
							self.boWindow.blur();
						}
						window.focus();
					}, 100);
				};
				self.boWindow = window.open(self.$contentUrl, "_blank");
			}
		} else {
			document.site.dom.empty(self.body);
			self.boWindow = null;
		}
	},
	renderLayoutContent: function() {
		VignetteField.prototype.renderLayoutContent.call(this);
		this.$contentUrl = this.$field.$location && this.$field.$location.$url;
		this._showUrl();
	},
	setFocus: function() {
		return true;
	},
	isDirty: function(value, record, metaData, $bind) {
		return false;
	},
	dispose: function() {
		window.sendAckCallTunnel = this.boWindow = null;
		VignetteField.prototype.dispose.call(this);
	},
	externalCall: function(proxy, values, opt) {
		var self = this,
			callDesc;

		function _resultCallback(res) {
			var err = !res ? [{
				message: "No result",
				severity: "error"
			}] : res.$diagnoses || null;

			err &&
				document.site.showDiagnoses({
					$diagnoses: err
				});
		}

		if (typeof proxy === 'object') {
			callDesc = proxy;
			proxy = undefined;
		}
		if (this.page && this.page.externalAdapter.onBlockExRpc) {
			this.page.externalAdapter.onBlockExRpc({
				field: opt && opt.attachedField || self,
				call: {
					proxy: proxy,
					callDesc: callDesc,
					values: values,
					callback: opt && opt.callback || _resultCallback
				}
			});
		}
	}

});
