// syracuse-ui/lib/page/landing/vignette/field/boWebIVignetteField.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var VignetteField = require("syracuse-ui/lib/page/landing/vignette/field/vignetteField").VignetteField;

var TUNNEL_DESC = {
	"spgm": "INTRUTILA:BOLINK",
	"params": [{ // FONCTION
		"dim": 1,
		"typ": "STRU",
		"siz": 8
	}, { // VARIA
		"dim": 1,
		"typ": "STRU",
		"siz": 8
	}, { // CLE
		"dim": 1,
		"typ": "STRU",
		"siz": 250
	}]
};

function BOWebIVignetteField() {}

exports.BOWebIVignetteField = helpers.defineClass(BOWebIVignetteField, VignetteField, {
	ensureDefaultTitle: function() {

	},
	setDataBind: function(value, record, metaData, $bind) {
		if (this.page.externalAdapter.setDataBind(this, value, record, metaData)) {
			if (value !== undefined) {
				this.$contentUrl = this.$item.$url = value;
				this._showUrl();
			}
		}
	},
	applyDesignMetaData: function(metaData, designing) {
		VignetteField.prototype.applyDesignMetaData.call(this, metaData, designing);
		if (metaData.$url !== undefined) {
			this.$contentUrl = this.$item.$url = metaData.$url;
		}
		if (metaData.$url !== undefined) {
			this._showUrl();
		}
	},
	_showUrl: function() {
		if (this.$contentUrl) {
			if (!this.boWindow) {
				var self = this;
				window.x3boSendAckCallTunnel = function(func, value, key) {
					setTimeout(function() {
						window.focus();
						self.externalCall(TUNNEL_DESC, [func, value, key]);
					}, 100);
				};
				this.boWindow = window.open(this.$contentUrl, "_blank");
			}
			this.domTitle.href = this.$contentUrl;
			this.domTitle.target = "_blank";
		} else {
			this.domTitle.href = "";
			this.domTitle.target = "";
			document.site.emptyDom(this.body);
			this.boWindow = null;
		}
		document.site.toggleClass(this.domTitle, "s-title-link", this.$contentUrl != null);
	},
	renderLayoutContent: function() {
		VignetteField.prototype.renderLayoutContent.call(this);
		this.$contentUrl = this.$field.$location && this.$field.$location.$url;
		var title = this.domTitle.textContent;
		this.domTitle.title = title;
		this.domTitle.textContent = title;
		this._showUrl();
	},
	setFocus: function() {
		return true;
	},
	isDirty: function(value, record, metaData, $bind) {
		return false;
	},
	dispose: function() {
		window.sendAckCallTunnel = this.boWindow = null;
		VignetteField.prototype.dispose.call(this);
	},
	externalCall: function(proxy, values, opt) {
		var self = this,
			callDesc;

		function _resultCallback(res) {
			var err = !res ? [{
				message: "No result",
				severity: "error"
			}] : res.$diagnoses || null;

			err &&
				document.site.showDiagnoses({
					$diagnoses: err
				});
		}

		if (typeof proxy === 'object') {
			callDesc = proxy;
			proxy = undefined;
		}
		if (this.page && this.page.externalAdapter.onBlockExRpc) {
			this.page.externalAdapter.onBlockExRpc({
				field: opt && opt.attachedField || self,
				call: {
					proxy: proxy,
					callDesc: callDesc,
					values: values,
					callback: opt && opt.callback || _resultCallback
				}
			});
		}
	}

});
