// syracuse-ui/lib/page/landing/vignette/field/navigationVignetteField.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var VignetteField = require("syracuse-ui/lib/page/landing/vignette/field/vignetteField").VignetteField;

function NavigationVignetteField() {}

exports.NavigationVignetteField = helpers.defineClass(NavigationVignetteField, VignetteField, {
	setDataBind: function(value, record, metaData, $bind) {

	},
	doPicker: function(pickerId) {
		VignetteField.prototype.doPicker.call(this, pickerId);
		switch (pickerId) {
			case "design":
				this.designArticle(!this.designer);
				break;
		}
	},
	designArticle: function(open) {
		document.site.dom.toggleClass(this.pickers.items.design, "s-close", open);
		document.site.designArticle(open, this, NavigationVignetteDesigner);
	},
	drawBox: function() {
		this.domItem = document.createElement("section");
		this.domItem.className = this.$skin;
		this.layoutSlot.appendChild(this.domItem);
		this.setState(this.$item);
		this.body = document.createElement("div");
		this.body.className = this.$skin + "-body";
		this.appendHeader();
		this.ensureLayoutMode();
		this.domItem.appendChild(this.body);
		this.applyDesignMetaData(this.$item, false);
		this.togglePicker("design", true);
		this.togglePicker("refresh", true);
		this.togglePicker("maximize", true);
		this.openBox(this.$item.$opened !== false);
	},
	dispose: function() {
		VignetteField.prototype.dispose.call(this);
	}
});
