// syracuse-ui/lib/page/landing/vignette/field/pageVignetteField.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var VignetteField = require("syracuse-ui/lib/page/landing/vignette/field/vignetteField").VignetteField;

function PageVignetteField() {}

exports.PageVignetteField = helpers.defineClass(PageVignetteField, VignetteField, {
	maximize: function() {
		VignetteField.prototype.maximize.call(this);
		if (this.vignette) {
			this.vignette.ensurePageVisibility();
		}
	},
	minimize: function() {
		VignetteField.prototype.minimize.call(this);
		if (this.vignette) {
			this.vignette.ensurePageVisibility();
		}
	},
	showAdvancedFields: function(show) {
		if (!this._isOnChildPageUpdate) {
			VignetteField.prototype.showAdvancedFields.call(this, show);
			if (this.vignette && this.vignette.showAdvancedFields) {
				this._isOnChildPageUpdate = true;
				this.vignette.showAdvancedFields(this.vignette.$isAdvancedFieldsVisible = this.$isAdvancedFieldsVisible);
				this._isOnChildPageUpdate = false;
			}
		}
	},
	doPicker: function(pickerId) {
		VignetteField.prototype.doPicker.call(this, pickerId);
		switch (pickerId) {
			case "design":
				this.designArticle(!this.designer);
				break;
		}
	},
	designArticle: function(open) {
		if (this.vignette && this.vignette.$facet == "$query") {
			switch (this.vignette.$facet) {
				case "$query":
				case "$cube":
					var fields = this.vignette.layoutContent.getFields();
					for (var ii = 0, jj = fields.length; ii < jj; ii++) {
						if (fields[ii].isList) {
							var open = !fields[ii].designer;
							fields[ii].designArticle(open);
							break;
						}
					}
					break;
			}
		}
	},
	renderLayoutContent: function($url, onVignetteRefresh) {
		var self = this;
		VignetteField.prototype.renderLayoutContent.call(self);
		if ($url) {
			self.$url = $url;
		}

		var category = this.$field && this.$field.$isLandingSearchVignette ? "landingSearchVignette" : "vignette";

		// to handle search vignette page (fk)
		if (!self.$url && self.$field && self.$field.$prototype) {
			self.vignette = document.site.loadNewPage({
				layoutSlot: self.body,
				$category: category,
				vignetteField: self,
				$facet: "$query",
				$autoFetch: false,
				$representation: self.$field
			});
			if (onVignetteRefresh) {
				self.onItemEnterLeave(null, true);
			}
		} else {
			document.controller.loadRepresentation(null, self.$url || self.$field.$location.$url, function($itemPage) {
				if (self.body) {
					$itemPage.layoutSlot = self.body;
					$itemPage.$category = category;
					$itemPage.vignetteField = self;
					if (!self.$url && self.$item.$vignette) {
						$itemPage.$representation.$article = self.$item.$vignette;
						delete self.$item.$vignette.$menus;
					}
					self.vignette = document.site.loadNewPage($itemPage);
					if (onVignetteRefresh) {
						self.onItemEnterLeave(null, true);
					}
				}
			});
		}
	},
	dispose: function() {
		if (this.vignette) {
			this.vignette.dispose();
			this.vignette = null;
		}
		VignetteField.prototype.dispose.call(this);
	}
});
