// syracuse-ui/lib/filter/view/gridRow.js
"use strict";
var _helpers = require('syracuse-core').helpers;

function GridCell(row, col) {}

_helpers.defineClass(GridCell, null, {
	load: function(row, col) {
		this.row = row;
		this.slot = syra_dom.td(row.filter_criteria_css);
		(this.col = col).filter_cell = this;
		if (!col.$field) {
			syra_dom.div(row.filter_criteria_css + "-empty", this.slot);
		} else {
			col.$isHidden && syra_dom.hide(this.slot, true);
			col.cssFilter = null;
			if (col.$field.$filterable) {
				this.field = syra_filter.addField(row, this.slot, {
					$field: col.$field,
					$bind: col.$item.$bind,
					$isTopLabelAlignment: false,
					$isGridCell: true,
					$inplace: true
				});
				col.cssRecordCellFilter = " s-filter-criteria-col";
				switch (col.$field.$type) {
					case "application/x-choice":
					case "application/x-date":
					case "application/x-time":
					case "application/x-reference":
						col.cssRecordCellFilter += " s-filter-criteria-col-with-btn";
				}

				// update filter cell if where params in url (issue #3022)
				var params = row.page.urlSeg && row.page.urlSeg.params;
				if (params && params.where) {
					var keys = [this.field.$filterBind = col.$item.$bind];
					// in case of reference field, appropriate $bind value is field.$reference.$value.$prop
					if (this.field.isReferenceField) {
						keys.push(this.field.$filterBind = this.field.$reference.$value.$prop);
					}
					var exp = syra_filter.criteria.parse(params.where, keys);
					if (exp && keys == keys) {
						syra_filter.operators.setFieldOperator(this.field, exp.$value, exp.$operator);
						if (this.field.isReferenceField) {
							var obj = {};
							obj[this.field.$filterBind] = exp.$value;
							exp.$value = obj;
						}
						this.field.setValue(exp.$value);
					}
				}
			}
		}
		this.append();
	},
	focus: function() {
		this.field.focus();
	},
	append: function() {
		this.row[this.col.table.$rowKey].appendChild(this.slot);
	},
	insertBefore: function(col) {
		var tr = this.row[this.col.table.$rowKey];
		tr.insertBefore(this.slot, (col && col.filter_cell && col.filter_cell.slot.nextSibling) || tr.firstChild);
	},
	hide: function(isHidden) {
		syra_dom.hide(this.slot, isHidden);
	},
	remove: function() {
		syra_dom.remove(this.slot);
		this.dispose();
	},
	dispose: function() {
		if (this.col) {
			delete this.col.filter_cell;
		}
		syra_article.dispose(this);
	}
});


function GridRow(list) {}

_helpers.defineClass(GridRow, null, {
	load: function(list) {
		this.list = list;
		this.filter_criteria_css = "s-filter-criteria-cell";
		this.isFilterArticle = this.isRecordArticle = true;
		this.$facet = "$filter";
		this.$isEditMode = true;
		this.$prototype = list.$prototype.$item;
		this.$prototype.$localization = list.page.$prototype.$localization;
		syra_item.initialize(list.page, this, {
			$layout: {
				$items: list.$item.$layout.$items
			}
		}, list);

		syra_article.beforeDraw(this);
		this.filterMap = {};
		list.builder.fixedTable && list.builder.fixedTable.headTable.appendChild(this.fixedRow = document.createElement("tr"));
		list.builder.scrollTable.headTable.appendChild(this.domItem = this.dataRow = document.createElement("tr"));
	},
	addCriteria: function(col) {
		(this.filterMap[col.key] = new GridCell()).load(this, col);
	},
	focus: function() {
		var map = this.list.builder.columnsMap;
		var $bind = this.list.builder.getNextFocusableField();
		$bind && map[$bind] && map[$bind].filter_cell && map[$bind].filter_cell.focus();
	},
	triggerFilter: function() {
		this.list.filler.trigger(this.list, {
			filter: syra_filter.getCriterion(this).join(" and ")
		});
		this.validateDisplay();
	},
	validateDisplay: function(event) {
		var show = syra_filter.getCriterion(this).length > 0;
		if (show && !this.filter_clear) {
			this.list.topbar.insertBefore(syra_filter.addClear(this).link, this.list.searcherSlot);
		}
		syra_button.hide(this.filter_clear, !show);
	},
	onFormUpdateDelta: function(field, value) {
		syra_filter.setCriteria(this, field, value);
		this.triggerFilter();
		return false;
	},
	dispose: function() {
		if (this.list) {
			this.filter_clear && syra_button.remove(this.filter_clear);
			var cols = this.list.builder.allColumns;
			if (cols) {
				for (var ii = 0, jj = cols.length; ii < jj; ii++) {
					cols[ii].filter_cell && cols[ii].filter_cell.dispose();
				}
			}
			delete this.list.filter_row;
		}
		syra_article.dispose(this);
	},
	onFieldFocusChange: function(field, isIn) {
		this.list.builder.scrollTable.switchHeadScroll(isIn);
	},
	setFilterable: function(filterable) {
		if (this.filterable !== filterable) {
			this.filterable = filterable;
			syra_dom.hide(this.fixedRow, !filterable);
			syra_dom.hide(this.dataRow, !filterable);
		}
	},

	fusion_getSortedFilterRcdFields: function() {
		return (Object.keys(this.filterMap)).sort(function(a, b) {
			return parseInt(a.substr(2), 10) - parseInt(b.substr(2), 10);
		});
	},
	fusion_getQuickSelValues: function() {
		var res = [];
		var $binds = this.fusion_getSortedFilterRcdFields();
		for (var ii = 0, jj = $binds.length; ii < jj; ii++) {
			var field = this.filterMap[$binds[ii]].field;
			field.$field && syra_filter.operators.setFieldOperator(field, field.getInputValue());
			res.push(field.getFilterCriteria() || "");
		}
		this.validateDisplay();
		return res;
	},
	fusion_getNextQuickSelField: function(bind, nextSel) {
		var ret = null,
			next, prev;
		if (this.$item && this.$item.$layout && this.$item.$layout.$items) {
			if (nextSel === undefined) {
				ret = this.boundFields[bind];
			} else {
				var cells = this._getSortedFilterRcdFields();
				for (var ii = 0, jj = cells.length; ii < jj; ii++) {
					if (cells[ii] === bind) {
						if (ii === 0) {
							next = cells[ii + 1];
							prev = bind;
						} else
						if (ii === (cells.length - 1)) {
							next = cells[0];
							prev = cells[ii - 1];
						} else {
							next = cells[ii + 1];
							prev = cells[ii - 1];
						}
						break;
					}
				}
				ret = this.boundFields[nextSel ? next : prev];
			}
		}
		return ret;
	}
});

exports.add = function(list) {
	if (list.filter_row) {
		list.filter_row.dispose();
	}
	(list.filter_row = new GridRow()).load(list);
};
