// syracuse-tablet/html/js/factories/appFactory.js
"use strict";

var log = require('syracuse-tablet/html/js/helpers/logger').log;
var utils = require('syracuse-tablet/html/js/helpers/utils');
var Application = require('syracuse-tablet/html/js/application/application').Application;
var DashoardPage = require('syracuse-tablet/html/js/pages/pageDashboard').Page;
var RegularPage = require('syracuse-tablet/html/js/pages/pageRegular').Page;
var HomePage = require('syracuse-tablet/html/js/pages/pageHome').Page;
var LoginPage = require('syracuse-tablet/html/js/pages/pageLogin').Page;
var dashboardCtrlr = require('syracuse-tablet/html/js/application/ctrlrDashboard').Controller;
var VignetteCtrlr = require('syracuse-tablet/html/js/application/ctrlrVignette').Controller;
var ViewCtrlr = require('syracuse-tablet/html/js/application/ctrlrView').Controller;
var DaoApp = require('syracuse-tablet/html/js/application/daoApp').Dao;
var DaoView = require('syracuse-tablet/html/js/application/daoView').Dao;

var _createApplication = function($elmt) {
	var res = window.smApp = new Application($elmt);
	res.init();
	log && log("Application", "Create ok");
	return res;
};

var _createPage = function($parent, state) {
	var res = null,
		type = state.type;
	if (type == "login") {
		res = new LoginPage($parent, state);
	} else if (type == "home") {
		res = new HomePage($parent, state);
	} else if (type == "dashboard") {
		res = new DashoardPage($parent, state);
	} else if (type == "regular") {
		res = new RegularPage($parent, state);
	} else {
		throw new Error("Unknown page type[" + type + "]");
	}
	log && log("New Page", "Create ok");
	return res;
};

var _createDashboardCtrlr = function(page) {
	var ctrlr = new dashboardCtrlr(page);
	log && log("New Dashboard controller", "Create ok");
	return ctrlr;
};

var _createVignetteCtrlr = function(ctrlrParent) {
	var ctrlr = new VignetteCtrlr(ctrlrParent);
	log && log("New Vignette controller", "Create ok");
	return ctrlr;
};

var _createViewCtrlr = function(page, type) {
	var ctrlr;
	switch (type) {
		case "representation":
		case "process":
		case "request":
		case "statistics":
		case "dashboardpage":
		case "externallink":
			ctrlr = new ViewCtrlr(page);
			break;
		default:
			throw new Exception("Unknown View type[" + type + "]");
			break;
	}
	log && log("New View controller", "Create ok");
	return ctrlr;
};

var _createDaoApp = function() {
	var dao = new DaoApp();
	log && log("New DAO Application", "Create ok");
	return dao;
};

var _createDaoView = function(type, data, prototype, parent, options) {
	var dao = new DaoApp();
	switch (type) {
		case "representation":
		case "process":
		case "request":
		case "statistics":
		case "dashboardpage":
		case "externallink":
			dao = new DaoView(data, prototype, parent, options);
			break;
		default:
			throw new Exception("Unknown View type[" + type + "]");
			break;
	}
	log && log("New DAO View", "Create ok");
	return dao;
};

exports.createApplication = _createApplication;
exports.createPage = _createPage;
exports.createDashboardCtrlr = _createDashboardCtrlr;
exports.createVignetteCtrlr = _createVignetteCtrlr;
exports.createViewCtrlr = _createViewCtrlr;
exports.createDaoApp = _createDaoApp;
exports.createDaoView = _createDaoView;
