// syracuse-tablet/html/js/factories/siteFactory.js
"use strict";

var log = require('syracuse-tablet/html/js/helpers/logger').log;
var utils = require('syracuse-tablet/html/js/helpers/utils');
var Site = require('syracuse-tablet/html/js/site/site').Site;
var DashoardPage = require('syracuse-tablet/html/js/site/pageDashboard').Page;
var RegularPage = require('syracuse-tablet/html/js/site/pageRegular').Page;
var dashboardCtrlr = require('syracuse-tablet/html/js/site/ctrlrDashboard').Controller;
var VignetteCtrlr = require('syracuse-tablet/html/js/site/ctrlrVignette').Controller;
var GadgetCtrlr = require('syracuse-tablet/html/js/site/ctrlrGadget').Controller;
var DaoSite = require('syracuse-tablet/html/js/site/daoSite').Dao;
var DaoGadget = require('syracuse-tablet/html/js/site/daoGadget').Dao;

var _createSite = function($elmt) {
	var res = window.smSite = new Site($elmt);
	log && log("Site", "Create ok");
	return res;
};

var _createPage = function(type, $parent, state) {
	var res = null;
	if (type == "dashboard") {
		res = new DashoardPage($parent, state);
	} else if (type == "regular") {
		res = new RegularPage($parent, state);
	} else {
		throw new Exception("Unknown page type[" + type + "]");
	}
	log && log("New Page", "Create ok");
	return res;
};

var _createDashboardCtrlr = function(page) {
	var ctrlr = new dashboardCtrlr(page);
	log && log("New Dashboard controller", "Create ok");
	return ctrlr;
};

var _createVignetteCtrlr = function(ctrlrParent) {
	var ctrlr = new VignetteCtrlr(ctrlrParent);
	log && log("New Vignette controller", "Create ok");
	return ctrlr;
};

var _createGadgetCtrlr = function(page, type) {
	var ctrlr;
	switch (type) {
		case "representation":
		case "process":
		case "request":
		case "statistics":
		case "dashboardpage":
		case "externallink":
			var ctrlr = new GadgetCtrlr(page);
			break;
		default:
			throw new Exception("Unknown Gadget type[" + type + "]");
			break;
	}
	log && log("New Gadget controller", "Create ok");
	return ctrlr;
};

var _createDaoSite = function() {
	var ctrlr = new DaoSite();
	log && log("New DAO Site", "Create ok");
	return ctrlr;
};

var _createDaoGadget = function(data, prototype, parent, options) {
	var ctrlr = new DaoGadget(data, prototype, parent, options);
	log && log("New DAO Gadget", "Create ok");
	return ctrlr;
};

exports.createSite = _createSite;
exports.createPage = _createPage;
exports.createDashboardCtrlr = _createDashboardCtrlr;
exports.createVignetteCtrlr = _createVignetteCtrlr;
exports.createGadgetCtrlr = _createGadgetCtrlr;
exports.createDaoSite = _createDaoSite;
exports.createDaoGadget = _createDaoGadget;
