// syracuse-ui/lib/authoring/articleTree.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');


function ArticleTree() {}

exports.ArticleTree = helpers.defineClass(ArticleTree, null, {
	load: function(awPalette, awArticle, slot) {
		this.awPalette = awPalette;
		slot.style.display = "none";
		document.site.emptyDom(slot);
		this._nodes = {};
		this._addNode(awArticle, {
			node: slot,
			isRoot: true
		});
		(this.slot = slot).style.display = "";
	},
	unload: function() {
		this.slot.style.display = "none";
		document.site.emptyDom(this.slot);
		this.dispose();
	},
	dispose: function() {
		if (this._nodes) {
			var ids = Object.keys(this._nodes);
			for (var ii = 0, jj = ids.length; ii < jj; ii++) {
				var node = this._nodes[ids[ii]];
				if (node) {
					node.item = node.node = node.children = node.data = null;
					node.opener = node.title = null;
				}
			}
		}
		this.selectedNode = this._nodes = this.slot = this.awPalette = null;
	},
	onNodeEvent: function(event) {
		var authorPage = document.site.authorPage;
		delete document.site.requestedDDAuthoringItem;
		event.stopPropagation();
		var node = this.findNode(event);
		if (node) {
			if (event.target.className.indexOf("s-aw-tree-item-opener-children") >= 0) {
				var open = !(node.opener.className.indexOf("s-open") < 0);
				node.children.style.display = open ? "" : "none";
				document.site.toggleClass(node.opener, "s-open", !open);
			} else {
				//event hidden  if card
				switch (event.type) {
					case "mouseenter":
						authorPage.toggleOverItem(node.item, true);
						this.awPalette.targetPage.scrollToItem(node.item);
						break;
					case "mouseleave":
						authorPage.toggleOverItem(node.item, false);
						break;
					case "click":
						if (node.item == this.awPalette.awArticle) {
							this.awPalette.targetPage.scrollToItem(node.item);
						}
						this.selectNode(node.item, true);
						this._isSelecteDisabled = true;
						authorPage.selectItem(node.item, true);
						this._isSelecteDisabled = false;
						break;
				}
			}
		}
	},
	findNode: function(event) {
		var target = event.target;
		while (target && !target.syraNodeId) {
			target = target.parentNode;
		}
		return (target && target.syraNodeId) ? this._nodes[target.syraNodeId] : null;
	},
	selectNode: function(item, select) {
		if (!this._isSelecteDisabled && item && !item.disposed) {
			var node = this._nodes[item.id];
			if (node) {
				if (select) {
					if (this.selectedNode && this.selectedNode != node && this.selectedNode.item) {
						document.site.toggleClass(this.selectedNode.title, "s-aw-tree-designed-" + this.selectedNode.item.$authoringLevel, false);
					}
					this.selectedNode = node;
					document.site.toggleClass(node.title, "s-aw-tree-designed-" + node.item.$authoringLevel, true);
					this.awPalette.scrollToItem(this.selectedNode.node, $(this.slot), true);
				} else {
					if (this.selectedNode == node) {
						this.selectedNode = null;
						document.site.toggleClass(node.title, "s-aw-tree-designed-" + node.item.$authoringLevel, false);
					}
				}
			}
		}
	},
	_addNode: function(item, parentNode) {
		var items;
		if (!item.isSpaceBox) {
			if (item.isLayout) {
				items = item.items;
			} else {
				//item.parent = parent;
				if (!parentNode.children) {
					parentNode.children = document.createElement("ul");
					parentNode.children.className = "s-aw-tree-level";
					parentNode.node.appendChild(parentNode.children);
					if (!parentNode.opener) {
						if (parentNode.title) {
							parentNode.opener = document.createElement("a");
							parentNode.opener.className = "s-aw-tree-item-opener";
							parentNode.title.parentNode.insertBefore(parentNode.opener, parentNode.title);
							parentNode.opener.className = "s-aw-tree-item-opener-children";
						}
					}
					if (parentNode.isRoot) {
						parentNode.children.className += " s-aw-tree-root";
					}
				}
				var itemNode = {};
				itemNode.item = item;
				itemNode.node = document.createElement("li");
				itemNode.node.syraNodeId = item.id;
				this._nodes[item.id] = itemNode;
				itemNode.node.className = "s-aw-tree-item s-aw-add-item";
				parentNode.children.appendChild(itemNode.node);

				itemNode.data = document.createElement("div");
				itemNode.data.className = "s-aw-tree-item-data";
				itemNode.node.appendChild(itemNode.data);

				itemNode.title = document.createElement("div");
				itemNode.title.className = "s-aw-tree-item-title s-aw-tree-" + item.$authoringLevel;
				if (this.awPalette.awItem == item) {
					itemNode.title.className += " s-aw-tree-designed-" + item.$authoringLevel;
				}
				var title = item.getTitle();
				if (title == "-" && item.getDefaultTitle) {
					title = item.getDefaultTitle();
				}
				var bindTitle = "",
					desc = title;
				if (item.$item && item.$item.$bind) {
					bindTitle = this.awPalette.getBindTitle(item.$item.$bind, item.$field);
					if (title == "") {
						desc = title = bindTitle;
					} else {
						desc = title + " (" + bindTitle + ")";
					}
				}
				itemNode.title.textContent = title;
				itemNode.node.title = desc;
				itemNode.data.appendChild(itemNode.title);
			}
			if (item.layoutContent) {
				items = item.layoutContent.items;
			}
			if (items) {
				for (var ii = 0, jj = items.length; ii < jj; ii++) {
					this._addNode(items[ii], itemNode || parentNode);
				}
			}
		}
	}
});
