// syracuse-ui/lib/authoring/authorPage.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var locale = require('syracuse-core/lib/locale');
var RawPage = require("syracuse-ui/lib/page/rawPage").RawPage;
var ItemPalette = require("./itemPalette").ItemPalette;
var DragDrop = require("./aside/dragDrop").DragDrop;
var RowResizer = require("./aside/rowResizer").RowResizer;

function AuthorPage(){
}

exports.AuthorPage = helpers.defineClass(AuthorPage, RawPage, {

    notifyUpdate: function(isUpdated){
        this.isUpdated = isUpdated !== false;
        this.palette.notifySteps();
    },
    _toggleAuthoringDialog: function(begin){
        var dialogWrapper = this.targetPage.dialogWrapper;
        if (dialogWrapper) {
            if (begin) {
                dialogWrapper.$$closeBtn[0].style.display = "none";
                dialogWrapper.overlay.className = "s-aw-overlay";
            }
            else {
                dialogWrapper.$$closeBtn[0].style.display = "";
                dialogWrapper.overlay.className = "s-overlay";
            }
        }
    },
    beginAuthoring: function(awPageSlot){
        this.preferences = document.site.ensurePreferences();
        this.preferences.authoring = this.preferences.authoring || {};
        this._initializePage();
        this.layoutSlot = awPageSlot;
        this.localize = locale.resources(module)();
        this.bindEvents();
        if (document.site.mainPage.isFusionPage) {
            this.targetPage = document.site.fusionGateway.activatedBook.selectedSheet;
        }
        else {
            this.targetPage = document.site.getTopDialogPage();
            if (this.targetPage) {
                this.targetPage = this.targetPage._content;
            }
        }
        this.diagnosePage = this.targetPage = this.targetPage || document.site.mainPage;
        if (this.targetPage.dialogWrapper) {
            this._toggleAuthoringDialog(true);
        }
        this.targetPage.authoringPage = this;
        this.targetPage.isOnAuthoring = true;
        if (this.targetPage.menuBar) {
            this.targetPage.menuBar.toggleItemAuthoring(true);
        }
        if (this.targetPage.fusionBar) {
            this.targetPage.fusionBar.toggleItemAuthoring(true);
        }
        this.paletteSlot = document.createElement("div");
        this.paletteSlot.className = "s-aw-palette-slot";
        this.targetPage._item.appendChild(this.paletteSlot);
        this.$sourceItem = helpers.object.clone(this.targetPage.$item, true);
        this.$defaultItem = helpers.object.clone(document.site.ensureDefaultArticle(this.targetPage.$prototype.$article, this.targetPage.$prototype), true);
        this.isUpdated = false;
        document.controller.loadWorkingCopy({
            menu: {
                $url: this.targetPage.$authorUrl
            },
            article: this,
            callback: function(){
                document.site.resize();
            }
        });
        this.rowResizer = new RowResizer();
        this.rowResizer.load();
    },
    onBeforeMainPageChange: function(continueChanging){
        var self = this;
        if (self.isUpdated) {
            document.site.showMessage({
                $title: self.localize.aw_updateMessageTitle,
                $message: self.localize.aw_cancelMessageText,
                $type: "question",
                $buttons: "yesnocancel",
                callback: function(response){
                    if (response.$id == "yes") {
                        self.menuItems.$save[0].click();
                        setTimeout(function(){
                            document.site.openAuthoringPage(false);
                            continueChanging();
                        }, 300);
                    }
                    else {
                        if (response.$id == "no") {
                            document.site.openAuthoringPage(false);
                            continueChanging();
                        }
                    }
                }
            });
            return false;
        }
        document.site.openAuthoringPage(false);
        return true;
    },
    onMenuClick: function(menuItem){
        var self = this;
        switch (menuItem.$item.$bind) {
            case "$delete":
                var selectedView = self.targetPage.pageViewSelector.getSelected();
                document.site.showMessage({
                    $title: self.localize.aw_deleteMessageTitle,
                    $message: self.localize.aw_deleteMessageText.replace("{0}", "'" + selectedView.$title + "'"),
                    $type: "question",
                    $buttons: "yesno",
                    callback: function(response){
                        if (response.$id == "yes") {
                            document.controller.executeMenu(menuItem);
                        }
                    }
                });
                return false;
                break;
            case "$close":
                if (self.isUpdated) {
                    document.site.showMessage({
                        $title: self.localize.aw_updateMessageTitle,
                        $message: self.localize.aw_cancelMessageText,
                        $type: "question",
                        $buttons: "yesnocancel",
                        callback: function(response){
                            if (response.$id == "yes") {
                                self.menuItems.$save[0].click();
                            }
                            else {
                                if (response.$id == "no") {
                                    self.targetPage.reloadLayout(helpers.object.clone(self.$sourceItem, true));
                                    document.site.openAuthoringPage(false);
                                }
                            }
                        }
                    });
                }
                else {
                    document.site.openAuthoringPage(false);
                }
                return false;
            default:
                if (menuItem.$sourceBind == "$aw_quit") {
                    this.targetPage.showDiagnoses(null);
                    document.site.openAuthoringPage(false);
                    return false
                }
                break;
        }
        if (menuItem.page != this && menuItem.page != this.palette) {
            if (!menuItem.page.isAuthoringFriend && !menuItem.boxParent.isDiagnoseMenus) {
                return !menuItem.page.vignetteField && !menuItem.page.isOnAuthoring;
            }
        }
        return true;
    },
    toggleViewType: function($viewType){
        this.$viewType = $viewType;
        if (this.isOnPreviewMode && $viewType != "preview") {
            this.isOnPreviewMode = false;
            this.paletteSlot.style.display = "";
            this.palette.toolsBar._slot.style.display = "";
            this.toggleItemAuthoring(this.targetPage, true);
            this.targetPage.onWindowResize();
        }
        switch ($viewType) {
            case "preview":
                if (!this.isOnPreviewMode) {
                    if (this.$isLightMode) {
                        this.$isLightMode = false
                        this.toggleItemAuthoring(this.targetPage, true);
                    }
                    this.isOnPreviewMode = true;
                    this.paletteSlot.style.display = "none";
                    this.palette.toolsBar._slot.style.display = "none";
                    this.toggleItemAuthoring(this.targetPage, false);
                    this.targetPage.onWindowResize();
                }
                break;
            case "structure":
                if (!this.$isLightMode) {
                    this.$isLightMode = true;
                    this.toggleItemAuthoring(this.targetPage, true);
                }
                break;
            case "layout":
                if (this.$isLightMode) {
                    this.$isLightMode = false
                    this.toggleItemAuthoring(this.targetPage, true);
                }
                break;
        }
    },
    drawBox: function(){
        document.site.emptyDom(this.layoutSlot);
        (this.$$item = $(this.layoutSlot)).show();
        this._paletteBar = document.createElement("div");
        this._paletteBar.setAttribute("id", "s-aw-page-left");
        this.layoutSlot.appendChild(this._paletteBar);
        
        var actions = document.createElement("div");
        actions.setAttribute("id", "s-aw-page-right");
        var close = this.loadNewItem(actions, {
            $bind: "$close",
            $title: this.localize.aw_close,
            $category: "link",
            $icon: {
                $mode: "icon"
            },
            $skin: "s-aw-page-close"
        });
        close.setMenu({
            $title: this.localize.aw_close
        }, null);
        this.loadNewItem(actions, {
            $layoutType: "row",
            $autoSize: true,
            $items: [{
                $category: "link",
                $bind: "$save",
                $skin: "s-aw-menus-save"
            }, {
                $category: "link",
                $bind: "saveAs",
                $skin: "s-aw-menus-saveas"
            }, {
                $category: "link",
                $bind: "$delete",
                $skin: "s-aw-menus-delete"
            }]
        });
        this.layoutSlot.appendChild(actions);
        (this.palette = new ItemPalette()).loadBox();
        
        
        document.site.resize();
        
        this.toggleItemAuthoring(this.targetPage, true);
        this.selectItem(this.targetPage, true);
    },
    selectItem: function(item, select, ensureSelect){
        if (this.awLayout && this.awLayout.disposed) {
            this.awLayout = null;
        }
        if (this.awItem && this.awItem.disposed) {
            this.awItem = null;
        }
        if (item && item.disposed) {
            item = this.awItem;
        }
        if (select) {
            item = item || this.targetPage;
            if (ensureSelect || item != this.awItem) {
                this.selectItem(null, false);
                if (item.isLayout) {
                    this.awItem = item.boxParent;
                    this.awLayout = item;
                }
                else {
                    this.awItem = item;
                    this.awLayout = null;
                }
                
                this.scrollToItem(this.awLayout || this.awItem);
                this.palette.showSettingPanel();
                if (this.awLayout) {
                    this._toggleItemCss(this.awLayout, "s-aw-designed", true);
                }
                this._toggleItemCss(item, "s-aw-designed", true);
                document.site.resize();
            }
        }
        else {
            item = item || this.awItem;
            if (item) {
                this._toggleItemCss(item, "s-aw-designed", false);
            }
            if (this.awLayout) {
                this._toggleItemCss(this.awLayout, "s-aw-designed", false);
            }
        }
        this.palette.toolsBar.treeContent.selectNode(item, select);
    },
    defineLayoutChildAuhtoringLevel: function(layout){
        if (layout.boxParent.$authoringLevel == "block") {
            layout.$chilAuthoringLevel = "field";
        }
        else {
            var children = this.getItemsFromLayout(layout);
            if (children.length > 0) {
                layout.$chilAuthoringLevel = children[0].$authoringLevel;
            }
            else {
                layout.$chilAuthoringLevel = layout.boxParent.$authoringLevel == "section" ? "block" : "section";
            }
        }
    },
    toggleDropArea: function(spaceBox, dropItem){
        if (dropItem) {
            if (spaceBox.deleteLink) {
                spaceBox.deleteLink.style.display = "none";
            }
            if (!spaceBox.dropArea) {
                spaceBox.dropArea = document.createElement("div");
                spaceBox.dropArea.className = "s-aw-drop-area-" + dropItem.$authoringLevel;
                spaceBox.dropArea.textContent = this.localize["aw_drop_" + dropItem.$authoringLevel];
                spaceBox._body.appendChild(spaceBox.dropArea);
            }
        }
        else {
            if (spaceBox.deleteLink) {
                spaceBox.deleteLink.style.display = "";
            }
            if (spaceBox.dropArea) {
                document.site.removeDomChild(spaceBox.dropArea);
                spaceBox.dropArea = null;
            }
        }
    },
    toggleItemAuthoring: function(item, enable){
        var site = document.site;
        if (item.fusionBar) {
            item.fusionBar.toggleItemAuthoring(enable || this.isOnPreviewMode, this.isOnPreviewMode);
        }
        if (item.menuBar) {
            item.menuBar.toggleItemAuthoring(enable || this.isOnPreviewMode, this.isOnPreviewMode);
        }
        if (!enable && item.loadPageViewSelector) {
            item.loadPageViewSelector();
        }
        if (item.articleParent && item.articleParent != this.targetPage) {
            return;
        }
        if (item.isMenuGroup || item.isMenuItem) {
            if (!this.targetPage.isDashBoard) {
                return;
            }
        }
        if ((item.$item && item.$item.$isAuthoringEnabled === false) ||
        (item.boxParent && item.boxParent.$item.$isAuthoringEnabled === false)) {
            return;
        }
        item.isAuthingEnabled = enable;
        if (item.isSpaceBox) {
            if (enable) {
                if (!item.deleteLink) {
                    if (item.boxParent.$authoringLevel == "block") {
                        item.$authoringLevel = "field";
                    }
                    else {
                        var children = item.boxParent.layoutContent ? this.getItemsFromLayout(item.boxParent.layoutContent) : [];
                        if (children.length > 0) {
                            item.$authoringLevel = children[0].$authoringLevel;
                        }
                        else {
                            item.$authoringLevel = item.boxParent.$authoringLevel == "section" ? "block" : "section";
                        }
                    }
                    item.deleteLink = document.createElement("a");
                    item.deleteLink.className = "s-aw-layout-delete";
                    item.deleteLink.textContent = this.localize.aw_deleteColumn;
                    item._body.appendChild(item.deleteLink);
                }
            }
            else {
                site.removeDomChild(item.deleteLink);
                item.deleteLink = null;
                site.toggleClass(item.layoutSlot, "s-aw-designed", false);
                site.toggleClass(item.layoutSlot, "s-aw-over", false);
            }
            site.toggleClass(item.layoutSlot, "s-aw-item", enable);
        }
        var domItem = item.dataSlot;
        if (!domItem && item.$$item) {
            domItem = item.$$item[0];
        }
        if (domItem) {
            //no domitem for layout
            if (item.tabTitle) {
                if (enable) {
                    item.tabTitle.setAttribute("data-s-item", item.id);
                }
                else {
                    site.toggleClass(item.tabTitle, "s-aw-designed", false);
                    site.toggleClass(item._body, "s-aw-designed", false);
                    site.toggleClass(item.tabTitle, "s-aw-over", false);
                    site.toggleClass(item._body, "s-aw-over", false);
                    item.tabTitle.removeAttribute("data-s-item");
                }
                site.toggleClass(item.tabTitle, "s-aw-item", enable);
                site.toggleClass(item._body, "s-aw-item", enable);
                site.toggleClass(item.tabTitle, "s-aw-light", enable && this.$isLightMode);
                site.toggleClass(item._body, "s-aw-light", enable && this.$isLightMode);
            }
            if (enable) {
                domItem.setAttribute("data-s-item", item.id);
                if (!item.isLayout) {
                    item.layoutSlot.setAttribute("data-s-item", item.id);
                }
            }
            else {
                if (domItem) {
                    if (item._body) {
                        site.toggleClass(item._body, "s-aw-over", false);
                        site.toggleClass(item._body, "s-aw-designed", false);
                    }
                    if (item._header) {
                        site.toggleClass(item._header, "s-aw-over", false);
                        site.toggleClass(item._header, "s-aw-designed", false);
                    }
                    site.toggleClass(item.layoutSlot, "s-aw-designed", false);
                    site.toggleClass(item.layoutSlot, "s-aw-over", false);
                    site.toggleClass(domItem, "s-aw-designed", false);
                    site.toggleClass(domItem, "s-aw-over", false);
                    domItem.removeAttribute("data-s-item");
                    item.layoutSlot.removeAttribute("data-s-item");
                }
            }
            if (item.$authoringLevel == "field") {
                item.layoutSlot.style.display = (this.$isLightMode || item.$isHidden) ? "none" : "";
            }
            if (item.isLayout) {
                if (enable) {
                    if (item.$layout.$layoutType == "row") {
                        site.toggleClass(item._item, "s-aw-item", true);
                        if (item.layoutSlot) {
                            site.toggleClass(item.layoutSlot, "s-aw-item", true);
                        }
                        if (this.$isLightMode) {
                            var children = this.getItemsFromLayout(item);
                            if (!children.length || children[0].$authoringLevel == "field") {
                                item._item.style.display = "none";
                            }
                        }
                        else {
                            item._item.style.display = "";
                        }
                    }
                }
                else {
                    site.toggleClass(domItem, "s-aw-item", false);
                    if (item.layoutSlot) {
                        site.toggleClass(item.layoutSlot, "s-aw-item", false);
                    }
                    item._item.style.display = "";
                }
            }
            else {
                if (item.layoutSlot) {
                    site.toggleClass(item.layoutSlot, "s-aw-item", enable);
                }
                site.toggleClass(domItem, "s-aw-item", enable);
                site.toggleClass(domItem, "s-aw-light", enable && this.$isLightMode);
                if (item.$isHidden) {
                    this.showHiddenItem(item);
                }
            }
        }
        if (item.rowSeparator) {
            site.toggleClass(item.rowSeparator, "s-aw-slot-sep-" + item.layoutParent.$layout.$layoutType, enable);
            item.rowSeparator.setAttribute(item.isLayout ? "data-s-layout" : "data-s-item", item.id);
        }
        if (item.isLayout) {
            if (item.items) {
                for (var ii = 0, jj = item.items.length; ii < jj; ii++) {
                    this.toggleItemAuthoring(item.items[ii], enable);
                }
            }
        }
        else {
            if (item.layoutContent) {
                this.toggleItemAuthoring(item.layoutContent, enable);
            }
        }
    },
    showHiddenItem: function(item){
        if (item.isAuthingEnabled) {
            var forced = item.$isHidden && !this.isOnPreviewMode;
            if (item.layoutSlot) {
                item.layoutSlot.style.display = (!item.$isHidden || (!this.$isLightMode && forced)) ? "" : "none";
                document.site.toggleClass(item.layoutSlot, "s-aw-visibility-forced", item.$isHidden);
            }
            var domItem = item._item || item._domItem;
            if (domItem) {
                domItem.style.display = (!item.$isHidden || (!this.$isLightMode && forced)) ? "" : "none";
                document.site.toggleClass(domItem, "s-aw-visibility-forced", item.$isHidden);
            }
        }
    },
    _toggleItemCss: function(item, css, show){
        if (item.$$item) {
            if (item.tabTitle) {
                document.site.toggleClass(item.tabTitle, css, show);
                document.site.toggleClass(item._body, css, show);
            }
            else {
                if (item.dataSlot) {
                    document.site.toggleClass(item.dataSlot, css, show); //page
                }
                else {
                    if (item.layoutSlot) {
                        document.site.toggleClass(item.layoutSlot, css, show);
                    }
                }
                if (item.isLayout) {
                    document.site.toggleClass(item._item, css, show);
                }
                if (item._header) {
                    document.site.toggleClass(item._header, css, show);
                }
                if (item._body) {
                    document.site.toggleClass(item._body, css, show);
                }
            }
        }
    },
    toggleOverItem: function(item, isOver){
        var css = "s-aw-over";
        this._toggleItemCss(item, css, isOver);
        if (isOver && this._overItem != item) {
            if (this._oveItems) {
                for (var ii = 0, jj = this._oveItems.length; ii < jj; ii++) {
                    this._toggleItemCss(this._oveItems[ii], css, false);
                }
            }
            this._oveItems = [];
            this._overItem = item;
            while (item) {
                this._toggleItemCss(item, css, true);
                this._oveItems.push(item);
                item = item.boxParent;
            }
        }
    },
    _onSaveAs: function(newData){
        if (newData.$actions.saveAs && newData.$actions.saveAs.$isDisabled && newData.$actions.saveAs.$isRequested === false) {
            if (newData.$authorUrl) {
                this.targetPage.$authorUrl = newData.$authorUrl;
            }
            if (newData.$actions.saveAs.$diagnoses) {
                for (var ii = 0, jj = newData.$actions.saveAs.$diagnoses.length; ii < jj; ii++) {
                    var $diagnose = newData.$actions.saveAs.$diagnoses[ii];
                    if (($diagnose.severity || $diagnose.$severity) !== "success") {
                        return false;
                    }
                }
            }
            delete newData.$actions.saveAs;
            this.notifyUpdate(false);
            this.targetPage.showDiagnoses({
                $diagnoses: [{
                    $message: this.localize.aw_saveMessageText,
                    $severity: "success"
                }],
                $links: {
                    $aw_quit: {
                        $title: this.localize.aw_quit
                    }
                }
            });
        }
    },
    _onSave: function(newData){
        if (newData.$actions.$save.$isDisabled && newData.$actions.$save.$isRequested === false) {
            if (newData.$authorUrl) {
                this.targetPage.$authorUrl = newData.$authorUrl;
            }
            if (newData.$actions.$save.$diagnoses) {
                for (var ii = 0, jj = newData.$actions.$save.$diagnoses.length; ii < jj; ii++) {
                    var $diagnose = newData.$actions.$save.$diagnoses[ii];
                    if (($diagnose.severity || $diagnose.$severity) !== "success") {
                        return false;
                    }
                }
            }
            newData.$actions.$save.$isDisabled = false;
            newData.$actions.$save.$isRequested = true;
            this.targetPage.showDiagnoses({
                $diagnoses: [{
                    $message: this.localize.aw_saveMessageText,
                    $severity: "success"
                }],
                $links: {
                    $aw_quit: {
                        $title: this.localize.aw_quit
                    }
                }
            });
        }
    },
    _onDelete: function(newData){
        if (newData.$actions.$delete.$isDisabled && newData.$actions.$delete.$isRequested === false) {
            if (newData.$authorUrl) {
                this.targetPage.$authorUrl = newData.$authorUrl;
            }
            if (newData.$actions.$delete.$diagnoses) {
                for (var ii = 0, jj = newData.$actions.$delete.$diagnoses.length; ii < jj; ii++) {
                    var $diagnose = newData.$actions.$delete.$diagnoses[ii];
                    if (($diagnose.severity || $diagnose.$severity) !== "success") {
                        return false;
                    }
                }
            }
            this.targetPage.showDiagnoses({
                $diagnoses: [{
                    $message: "Deleted Ok",
                    $severity: "success"
                }]
            });
            this.menuItems["$aw_quit"][0].click();
            
        }
    },
    applyChange: function(newData){
        if (newData && newData.$actions) {
            if (newData.$actions.saveAs) {
                this._onSaveAs(newData);
            }
            if (newData.$actions.$save) {
                this._onSave(newData);
            }
            if (newData.$actions.$delete) {
                this._onDelete(newData);
            }
        }
        RawPage.prototype.applyChange.call(this, newData);
    },
    _temporalySave: function(){
        var content = this.ensureSendBag().content = {
            $article: this.targetPage.$item
        };
        if (this.targetPage.saveGarbageLinks) {
            this.targetPage.saveGarbageLinks(content.$article);
        }
    },
    notifyActionChange: function(target, value, notifyServer){
        this._temporalySave();
        RawPage.prototype.notifyActionChange.call(this, target, value, notifyServer);
    },
    notifyDataChange: function(field, value){
        this._temporalySave();
        RawPage.prototype.notifyDataChange.call(this, field, value);
    },
    addNewSection: function(layoutParent, dropItem, index, $newCategory){
        layoutParent.removeSpaceBox();
        var $item = null;
        if (!dropItem) {
            $item = {
                $applyDefaultTitle: true
            };
            if ($newCategory != "menus") {
                if (this.targetPage.isDashBoard) {
                    if (layoutParent.boxParent.isSection) {
                        $newCategory = "menus";
                    }
                }
            }
            if ($newCategory == "menus") {
                $item.$category = $newCategory;
                $item.$isHidden = false;
            }
            else {
                $item.$category = "section";
            }
        }
        var newSection = layoutParent.loadChildItem(dropItem, $item, index);
        if (!newSection.layoutContent) {
            newSection.openBox(true);
        }
        return newSection;
    },
    wrapLayoutContent: function(layoutContent){
        var $newLayout = {
            $layoutType: layoutContent.$layout.$layoutType,
            $widths: layoutContent.$layout.$widths,
            $autoSize: layoutContent.$layout.$autoSize
        };
        var items = layoutContent.items.slice(0);
        for (var ii = 0, jj = items.length; ii < jj; ii++) {
            layoutContent.extractItem(items[ii]);
        }
        this.newConvert(layoutContent, {
            $layoutType: "stack"
        });
        var prevLayoutContent = layoutContent.createChildItem($newLayout);
        prevLayoutContent.removeSpaceBox(true);
        for (var ii = 0, jj = items.length; ii < jj; ii++) {
            prevLayoutContent.loadChildItem(items[ii], null, ii);
        }
        prevLayoutContent.$chilAuthoringLevel = layoutContent.$chilAuthoringLevel;
        delete layoutContent.$chilAuthoringLevel;
        return prevLayoutContent;
    },
    addSiblingToLayout: function(targetLayout, $insertAt){
        if (!targetLayout.layoutParent) {
            return this.addSiblingToLayout(this.wrapLayoutContent(targetLayout), $insertAt);
        }
        var $newLayout = {
            $layoutType: targetLayout.$layout.$layoutType,
            $widths: targetLayout.$layout.$widths,
            $autoSize: targetLayout.$layout.$autoSize
        };
        var targetItemIndex = targetLayout.layoutParent.getItemIndex(targetLayout, true);
        if ($insertAt == "bottom") {
            targetItemIndex++;
        }
        return targetLayout.layoutParent.loadChildItem(null, $newLayout, targetItemIndex);
    },
    insertFieldsWrapper: function(targetItem, $insertAt, $newCategory){
        var newChild;
        var items = targetItem.layoutContent.items.slice(0);
        if (this.hasLevelItems(items, "field")) {
            for (var ii = 0, jj = items.length; ii < jj; ii++) {
                targetItem.layoutContent.extractItem(items[ii]);
            }
            var $prevLayoutType = {
                $layoutType: targetItem.layoutContent.$layout.$layoutType,
                $widths: targetItem.layoutContent.$layout.$widths,
                $autoSize: targetItem.layoutContent.$layout.$autoSize
            };
            this.newConvert(targetItem.layoutContent, {
                $layoutType: "stack"
            });
            newChild = this.addNewSection(targetItem.layoutContent, null, 0);
            this.newConvert(newChild.layoutContent, $prevLayoutType);
            newChild.layoutContent.removeSpaceBox(true);
            for (var ii = 0, jj = items.length; ii < jj; ii++) {
                newChild.layoutContent.loadChildItem(items[ii], null, ii);
            }
        }
        else {
        
            newChild = this.addNewSection(targetItem.layoutContent, null, 0, $newCategory);
        }
        return newChild;
    },
    onAddNewField: function(targetItem, $newItem, $insertAt){
        if ($newItem.$fieldCategory == "link") {
            delete this.targetPage.garbageLinks[$newItem.$fieldBind];
        }
        var newChild;
        if ($insertAt) {
            newChild = this.targetPage.layoutContent.loadChildItem(null, {
                $bind: $newItem.$fieldBind,
                $category: $newItem.$fieldCategory
            });
            newChild.layoutParent.extractItem(newChild);
        }
        newChild = this.addNewItem(targetItem, "field", $newItem.$fieldBind, newChild, $insertAt, $newItem.$fieldCategory);
        this.palette.toolsBar.onExcludeField($newItem.$fieldBind, true);
        return newChild;
    },
    addNewItem: function(targetItem, $newAuthoringLevel, $bind, dropItem, $insertAt, $newCategory){
        var newChild;
        var targetBoxParent = targetItem.boxParent;
        var targetItemIndex = targetItem.layoutParent ? targetItem.layoutParent.getItemIndex(targetItem.$item) : undefined;
        
        if ($insertAt !== "undefined") {
            if (targetItem.layoutParent) {
                switch (targetItem.layoutParent.$layout.$layoutType) {
                    case "stack":
                        switch ($insertAt) {
                            case "top":
                                break;
                            case "bottom":
                                targetItemIndex++;
                                break;
                            case "left":
                            case "right":
                                /*if (targetItem.layoutParent.layoutParent && (targetItem.layoutParent.layoutParent.$layout.$layoutType == "row")) {
                         return this.addNewItem(targetItem.layoutParent, $newAuthoringLevel, $bind, dropItem, $insertAt);
                         }*/
                                if (!(targetItem.isLayout && targetItem.$layout.$layoutType == "row")) {
                                    var rowLayout = targetItem.layoutParent.createChildItem({
                                        $layoutType: "row",
                                        $items: []
                                    }, null, targetItemIndex);
                                    rowLayout.loadChildItems([targetItem]);
                                    rowLayout.$layout.$widths = null;
                                    targetItemIndex = $insertAt == "left" ? 0 : 1;
                                }
                                else {
                                    targetItem.loadChildItem(dropItem, null, $insertAt == "left" ? 0 : targetItem.items.length);
                                    return dropItem;
                                }
                                break;
                        }
                        break;
                    case "row":
                        switch ($insertAt) {
                            case "top":
                            case "bottom":
                                //drop in row empty cell, wrap drop in stack
                                targetItem.layoutParent.wrapIntack([targetItem], targetItemIndex);
                                targetItemIndex = ($insertAt == "top") ? 0 : 1;
                                break;
                            case "left":
                            case "right":
                                targetItem.layoutParent.$layout.$widths = null;
                                if ($insertAt == "right") {
                                    targetItemIndex++;
                                }
                                break;
                        }
                        break;
                    case "tabs":
                        if ($insertAt == "right") {
                            targetItemIndex++;
                        }
                        break;
                }
            }
            else {
                targetItemIndex = 0;
            }
        }
        if (dropItem && dropItem.isLayout) {
            if (targetItem.isLayout && !targetItem.layoutParent) {
                targetItem = this.wrapLayoutContent(targetItem);
            }
            newChild = targetItem.layoutParent.loadChildItem(dropItem, null, targetItemIndex);
        }
        else {
            if (targetItem.isLayout && !targetItem.layoutParent) {
                targetItem = this.wrapLayoutContent(targetItem);
            }
            if (targetItem.isLayout) {
                this.defineLayoutChildAuhtoringLevel(targetItem);
            }
            
            switch (targetItem.$chilAuthoringLevel || targetItem.$authoringLevel) {
                case "article":
                    if ($newAuthoringLevel == "field") {
                        newChild = this.addNewItem(targetItem, "section");
                        newChild = this.addNewItem(newChild, $newAuthoringLevel, $bind, dropItem);
                    }
                    else {
                        if (!dropItem) {
                            newChild = this.insertFieldsWrapper(targetItem);
                            if ($newAuthoringLevel == "block") {
                                newChild = this.insertFieldsWrapper(newChild, undefined, $newCategory);
                            }
                        }
                        else {
                            newChild = this.addNewSection(targetItem.layoutContent, dropItem, 0);
                        }
                    }
                    break;
                case "section":
                    switch ($newAuthoringLevel) {
                        case "field":
                            if (dropItem) {
                                newChild = this.addNewSection(targetItem.layoutParent, null, targetItemIndex);
                                newChild = newChild.layoutContent.loadChildItem(dropItem);
                            }
                            else {
                                var children = this.getItemsFromLayout(targetItem.layoutContent);
                                if (children.length == 0) {
                                    newChild = targetItem.layoutContent.loadChildItem(null, {
                                        $bind: $bind
                                    });
                                }
                                else {
                                    if (children[0].$authoringLevel == "field") {
                                        newChild = children[0].layoutParent.loadChildItem(null, {
                                            $bind: $bind
                                        }, children[0].layoutParent.getItemIndex(children[0].$item));
                                    }
                                    else {
                                        newChild = this.addNewItem(targetItem, children[0].$authoringLevel);
                                        newChild = newChild.layoutContent.loadChildItem(null, {
                                            $bind: $bind
                                        });
                                    }
                                }
                            }
                            break;
                        case "section":
                            newChild = this.addNewSection(targetItem.layoutParent, dropItem, targetItemIndex);
                            break;
                        case "block":
                            if (dropItem) {
                                newChild = targetItem.layoutParent.loadChildItem(dropItem, null, targetItemIndex);
                            }
                            else {
                                newChild = this.insertFieldsWrapper(targetItem, undefined, $newCategory);
                            }
                            break;
                    }
                    break;
                case "block":
                    switch ($newAuthoringLevel) {
                        case "field":
                            if (dropItem) {
                                newChild = this.addNewSection(targetItem.layoutParent, null, targetItemIndex);
                                newChild = newChild.layoutContent.loadChildItem(dropItem);
                            }
                            else {
                                newChild = targetItem.layoutContent.loadChildItem(null, {
                                    $bind: $bind
                                });
                            }
                            break;
                        case "section":
                            if (dropItem) {
                                var children = dropItem.layoutContent.items.slice(0);
                                if (this.hasLevelItems(children, "block")) {
                                    targetItem.layoutParent.loadChildItems(children, targetItemIndex);
                                    targetItem.layoutParent.removeItem(dropItem, true);
                                    newChild = targetItem;
                                }
                                else {
                                    newChild = targetItem.layoutParent.loadChildItem(dropItem, null, targetItemIndex);
                                }
                            }
                            else {
                                newChild = this.addNewItem(targetItem.boxParent, $newAuthoringLevel);
                            }
                            break;
                        case "block":
                            newChild = this.addNewSection(targetItem.layoutParent, dropItem, targetItemIndex, $newCategory);
                            break;
                    }
                    break
                case "field":
                    var sectionParent = targetBoxParent;
                    if (targetBoxParent.$authoringLevel == "block") {
                        sectionParent = targetBoxParent.boxParent; //else $authoringLevel== section or article = no need of boxparent
                    }
                    switch ($newAuthoringLevel) {
                        case "section":
                        case "block":
                            if (dropItem) {
                                //dragged section							
                                var children = this._extractFields(dropItem.layoutContent);
                                targetItem.layoutParent.loadChildItems(children, targetItemIndex);
                                targetItem.layoutParent.removeItem(dropItem, true);
                                newChild = targetItem;
                            }
                            else {
                                var layoutParent, newTopChild;
                                if ($newAuthoringLevel == "block") {
                                    layoutParent = sectionParent.layoutContent;
                                    if (targetItem.boxParent.$authoringLevel == "block") {
                                        targetItemIndex = targetItem.boxParent.layoutParent.getItemIndex(targetItem.boxParent.$item);
                                    }
                                }
                                else {
                                    if (sectionParent.layoutParent) {
                                        layoutParent = sectionParent.layoutParent;
                                        targetItemIndex = layoutParent.getItemIndex(sectionParent.$item);
                                    }
                                    else {
                                        layoutParent = sectionParent.layoutContent;
                                        targetItemIndex = undefined;
                                    }
                                }
                                var children = layoutParent.items.slice(0);
                                newTopChild = newChild = this.addNewSection(layoutParent, null, targetItemIndex);
                                if (newTopChild.$authoringLevel == "section" && $newAuthoringLevel == "block") {
                                    newChild = this.addNewSection(newTopChild.layoutContent);
                                }
                                if (targetItem.boxParent == newTopChild.boxParent) {
                                    newChild.layoutContent.loadChildItems(children);
                                }
                            }
                            break;
                        case "field":
                            var $item = null;
                            if (!dropItem) {
                                $item = {
                                    $bind: $bind
                                };
                                if ($newCategory == "link") {
                                    $item.$category = $newCategory;
                                    $item.$skin = "s-mn-h2-link";
                                }
                            }
                            newChild = targetItem.layoutParent.loadChildItem(dropItem, $item, targetItemIndex);
                            break;
                            
                    }
                    break;
            }
        }
        if (!newChild) {
            newChild = targetItem && targetItem.isSpaceBox;
            newChild = newChild || this.targetPage;
        }
        
        if (targetItem && targetItem.isSpaceBox) {
            targetItem.layoutParent.removeItem(targetItem, true, false);
        }
        return newChild;
    },
    
    _notifyChildFields: function(parentItem, value, binding){
        var items = parentItem.layoutContent ? parentItem.layoutContent.items : parentItem.items;
        if (items) {
            for (var ii = 0, jj = items.length; ii < jj; ii++) {
                var childItem = items[ii];
                if (childItem.$authoringLevel == "field") {
                    var metaData = {}
                    metaData[binding] = value;
                    if (childItem.applyDesignMetaData) {
                        //not available for menu
                        childItem.applyDesignMetaData(metaData, true);
                    }
                }
                else {
                    this._notifyChildFields(childItem, value, binding);
                }
            }
        }
    },
    _notifyChildBox: function(parentItem, metaData){
        var items = parentItem.layoutContent ? parentItem.layoutContent.items : parentItem.items;
        parentItem.applyDesignMetaData(metaData, true);
        if (parentItem.layoutContent) {
            for (var ii = 0, jj = items.length; ii < jj; ii++) {
                var childItem = items[ii];
                if (childItem.$authoringLevel != "field") {
                    this._notifyChildBox(childItem, metaData);
                }
            }
        }
    },
    findTargetItemEvent: function(target, findLayout){
        var item;
        if (findLayout || target.getAttribute("data-s-layout")) {
            item = document.controller.findLayout($(target));
        }
        else {
            item = document.controller.findItem($(target));
        }
        return item;
    },
    authorizeEvent: function(item){
        return (item.page && (item.page.$dialogMode || (item.page == this) || (item.page == this.palette) || (item.page == this.palette.toolsBar)));
    },
    bindEvents: function(){
        var self = this;
        document.site.$$layoutSlot.delegate(".s-aw-item", "click.author", function(event){
            self.onClickItem(self.findTargetItemEvent(this), this);
            return false;
        }).delegate(".s-aw-item, .s-aw-add-item", "mousedown.author", function(event){
            if (document.site.dragDropInstance) {
                if (!document.site.dragDropInstance.isDragAuthoring) {
                    return;
                }
                document.site.dragDropInstance.stop(event);
                document.site.setDragDropInstance();
            }
            else {
                if (event.currentTarget.className.indexOf("s-aw-add-item") >= 0) {
                    if (event.currentTarget.className.indexOf("s-aw-tree-item") >= 0) {
                        var node = self.palette.toolsBar.treeContent.findNode(event);
                        if (node) {
                            document.site.requestedDDAuthoringItem = node.item;
                            return false;
                        }
                    }
                    else {
                        document.site.requestedDDAuthoringItem = document.controller.findItem($(event.currentTarget));
                        return false;
                    }
                }
                else {
                    var item = self.findTargetItemEvent(this);
                    if (item) {
                        if (item != self.targetPage && item.page == self.targetPage) {
                            document.site.requestedDDAuthoringItem = item;
                            return false;
                        }
                    }
                }
            }
            // return false;
        }).delegate(".s-aw-slot-sep-row,.s-aw-slot-sep-stack", "mousemove.author", function(event){
            if (!document.site.dragDropInstance) {
                var item = self.findTargetItemEvent(this);
                if (item) {
                    if (item != self.targetPage) {
                        if (document.site.requestedDDAuthoringItem) {
                            (document.site.dragDropInstance = new DragDrop()).start(self.targetPage.$$scrollview);
                        }
                        else {
                            if (document.site.dragDropInstance) {
                                if (item.layoutParent || item.isLayout) {
                                    document.site.dragDropInstance.moveDragImage(event, item);
                                }
                            }
                        }
                    }
                    self.toggleOverItem(item, true);
                }
                return false;
            }
            return true;
        }).delegate(".s-aw-item,.s-aw-add-item", "mousemove.author", function(event){
            if (!document.site.dragDropInstance || document.site.dragDropInstance.isDragAuthoring) {
                if (event.currentTarget.className.indexOf("s-aw-add-item") >= 0) {
                    if (event.currentTarget.className.indexOf("s-aw-tree-item") >= 0) {
                        var node = self.palette.toolsBar.treeContent.findNode(event);
                        if (node) {
                            var item = node.item;
                            if (document.site.requestedDDAuthoringItem) {
                                document.site.setDragDropInstance(new DragDrop());
                                document.site.dragDropInstance.start(self.targetPage.$$scrollview);
                            }
                            else {
                                if (document.site.dragDropInstance) {
                                    if (item.layoutParent || item.isLayout) {
                                        document.site.dragDropInstance.moveDragImage(event, item);
                                    }
                                }
                            }
                        }
                    }
                    else {
                        if (document.site.requestedDDAuthoringItem) {
                            var menuItem = document.site.requestedDDAuthoringItem;
                            var $drop = {
                                $awAddNewItem: true
                            };
                            switch (menuItem.$item.$bind) {
                                case "$addSection":
                                    $drop.$authoringLevel = "section";
                                    break;
                                case "$addBlock":
                                    $drop.$authoringLevel = "block";
                                    break;
                                case "$addMenus":
                                    $drop.$authoringLevel = "block";
                                    $drop.$category = "menus";
                                    break;
                                default:
                                    $drop.$authoringLevel = "field";
                                    $drop.$fieldBind = menuItem.$item.$fieldBind;
                                    $drop.$fieldCategory = $drop.$category = menuItem.$item.$fieldCategory;
                                    break;
                            }
                            $drop.$title = menuItem.getTitle();
                            document.site.requestedDDAuthoringItem = $drop;
                            document.site.setDragDropInstance(new DragDrop());
                            document.site.dragDropInstance.start(self.targetPage.$$scrollview);
                        }
                    }
                }
                else {
                    var item = self.findTargetItemEvent(this);
                    if (item) {
                        if (item != self.targetPage) {
                            if (document.site.requestedDDAuthoringItem) {
                                document.site.setDragDropInstance(new DragDrop());
                                document.site.dragDropInstance.start(self.targetPage.$$scrollview);
                            }
                            else {
                                if (document.site.dragDropInstance) {
                                    if (item.layoutParent || item.isLayout) {
                                        document.site.dragDropInstance.moveDragImage(event, item);
                                    }
                                }
                            }
                        }
                        self.toggleOverItem(item, true);
                    }
                }
                return false;
            }
            return true;
        }).delegate(".s-aw-item", "mouseenter.author mouseleave.author", function(event){
            if (document.site.dragDropInstance && document.site.dragDropInstance.isDragAuthoring) {
                var $$item = $(this);
                var layout = self.findTargetItemEvent(this, true);
                if (layout && layout.$layout.$layoutType == "tabs") {
                    if (event.type == "mouseenter") {
                        var tabItem = self.findTargetItemEvent(this);
                        self._ddTabEnter = setTimeout(function(){
                            if (tabItem && tabItem.$item) {
                                if (!tabItem.$item.$opened) {
                                    tabItem.openBox(true);
                                }
                            }
                        }, 500);
                    }
                    else {
                        if (self._ddTabEnter) {
                            clearTimeout(self._ddTabEnter);
                            delete self._ddTabEnter;
                        }
                    }
                }
                return false;
            }
            return true;
            // return false;
        }).delegate(".s-aw-layout-delete", "click.author", function(){
            if (!document.site.dragDropInstance) {
                delete document.site.requestedDDAuthoringItem;
                var spaceBox = document.controller.findItem($(this));
                setTimeout(function(){
                    self.deleteLayout(spaceBox);
                    self.endTargetPageUpdate(null, true);
                }, 10);
            }
            return false;
        });
    },
    onClickItem: function(item, target){
        var article = document.controller.findArticle($(target));
        if (article.isOnAuthoring) {
            if (document.site.dragDropInstance && document.site.dragDropInstance.isDragAuthoring) {
                document.site.dragDropInstance.stop(event);
                document.site.setDragDropInstance();
            }
            else {
                delete document.site.requestedDDAuthoringItem;
                item = item || document.controller.findItem($(target));
                if (item) {
                    if (item.isSpaceBox) {
                        item = item.boxParent;
                    }
                    if (item.isMenuGroup && item.contextField) {
                        return false;
                    }
                    if (item && (item != this.palette && item.page != this.palette)) {
                        this.selectItem(item, true, true);
                    }
                }
            }
        }
        else {
            if (article.page != this && article.page != this.palette) {
                if (!(article.page.$prototype && article.page.$prototype.$representation == "authoringSaveParam")) {
                    return false;
                }
            }
        }
        return true;
    },
    _extractFields: function(layout, fields, onlyDirectChild){
        fields = fields || [];
        if (layout) {
            for (var ii = layout.items.length - 1; ii >= 0; ii--) {
                var item = layout.items[ii];
                if (item != layout.boxParent) {
                    if (item.$authoringLevel === "field") {
                        fields.push(item);
                    }
                    else {
                        if (onlyDirectChild && !item.isLayout) {
                            continue;
                        }
                        fields = this._extractFields(item.isLayout ? item : item.layoutContent, fields);
                    }
                }
            }
        }
        return fields;
    },
    hasLevelItems: function(items, $authoringLevel){
        var res = false;
        if (items) {
            for (var ii = 0, jj = items.length; ii < jj && !res; ii++) {
                var item = items[ii];
                if (item.$authoringLevel == $authoringLevel && !item.isSpaceBox) {
                    res = true;
                }
                else 
                    if (item.isLayout) {
                        res = this.hasLevelItems(item.items, $authoringLevel);
                    }
            }
        }
        return res;
    },
    getItemsFromLayout: function(layout, children, $authoringLevel){
        if (!children) {
            children = [];
        }
        for (var ii = 0, jj = layout.items.length; ii < jj; ii++) {
            var item = layout.items[ii];
            if (item != layout.boxParent) {
                if (item.isLayout) {
                    children = this.getItemsFromLayout(item, children, $authoringLevel);
                }
                else {
                    if (!$authoringLevel || ($authoringLevel == item.$authoringLevel)) {
                        children.push(item);
                    }
                }
            }
        }
        return children;
    },
    getItemSettings: function($layout, $items){
        if (!$items) {
            $items = [];
        }
        if ($layout) { //null for spacebox
            for (var ii = 0, jj = $layout.$items.length; ii < jj; ii++) {
                var $item = $layout.$items[ii];
                if ($item.$items && !$item.$bind && !$item.$category) {
                    if (!$item.$layoutType) {
                        $item.$layoutType = "stack";
                    }
                    this.getItemSettings($item, $items);
                }
                else {
                    $items.push($item);
                }
            }
        }
        return $items;
    },
    _deleteLayoutRowItem: function(layout, deletedIndex){
        var widths = this.layoutValidator.getWidthValues(layout.$layout.$widths);
        var deletedWidth = widths[deletedIndex];
        var updatedIndex = Math.min(deletedIndex > 0 ? (deletedIndex - 1) : 1, widths.length - 1);
        widths[updatedIndex] += deletedWidth;
        widths.splice(deletedIndex, 1);
        layout.$layout.$widths = widths.join(",");
        if (layout.items.length == 1) {
            this.newConvert(layout, {
                $layoutType: "stack",
                $width: "100"
            });
        }
    },
    deleteLayout: function(spaceBox){
        var layout = spaceBox.layoutParent;
        var layoutParent = layout.layoutParent;
        var deletedIndex;
        if (spaceBox) {
            deletedIndex = layout.getItemIndex(spaceBox, true);
            layout.removeItem(spaceBox, true);
        }
        if (layout.items.length > 0 && layout.$layout.$layoutType == "row") {
            this._deleteLayoutRowItem(layout, deletedIndex);
        }
        else {
            if (layoutParent) {
                layoutParent.removeItem(layout, true);
                if (layoutParent.$layout.$layoutType == "row") {
                    this._deleteLayoutRowItem(layoutParent, deletedIndex);
                }
            }
            else {
                layoutParent = layout.boxParent.layoutParent;
                layoutParent.removeItem(layout.boxParent, true, false);
            }
        }
    },
    endTargetPageUpdate: function(selectedItem, isStructureUpdated){
        this.layoutValidator.validate(this.targetPage.layoutContent, true);
        if (isStructureUpdated) {
            this.palette.toolsBar.treeContent.toggle(true);
        }
        this.selectItem(selectedItem || this.awItem, true, true);
        this.notifyUpdate();
        this._temporalySave();
    },
    onDropItem: function(targetItem, dropItem, $insertAt, $drag){
        var newChild;
        if ($drag.isTab) {
            if ($drag.isTabBody || ($insertAt == "bottom" || $insertAt == "top")) {
                var isDropItemTab = dropItem.isTabLayout;
                var layoutRoot = targetItem.layoutParent;
                if (!targetItem.layoutParent.layoutParent) {
                    var children = this.getItemsFromLayout(layoutRoot);
                    for (var ii = 0, jj = children.length; ii < jj; ii++) {
                        layoutRoot.extractItem(children[ii]);
                    }
                    this.newConvert(layoutRoot, {
                        $layoutType: "stack"
                    });
                    var layoutTab = layoutRoot.loadChildItem(null, {
                        $layoutType: "tabs"
                    });
                    for (var ii = 0, jj = children.length; ii < jj; ii++) {
                        if (dropItem != children[ii]) {
                            layoutTab.loadChildItem(children[ii]);
                        }
                    }
                    layoutRoot = layoutTab;
                }
                targetItem = layoutRoot;
                /*if (dropItem.$awAddNewItem) {
                 if (dropItem.$fieldBind) {
                 newChild = this.onAddNewField(layoutRoot, dropItem, $insertAt);
                 }
                 else {
                 if ($insertAt) {
                 newChild = this.addNewSection(this.targetPage.layoutContent, null, 0, dropItem.$category);
                 newChild.layoutParent.extractItem(newChild);
                 }
                 newChild = this.addNewItem(layoutRoot, dropItem.$authoringLevel, null, newChild, $insertAt, dropItem.$category);
                 }
                 }
                 else {
                 newChild = this.addNewItem(layoutRoot, dropItem.$authoringLevel, dropItem.$bind, dropItem, $insertAt);
                 }*/
            }
        }
        if (!newChild) {
            if (dropItem.$awAddNewItem) {
                if (dropItem.$fieldBind) {
                    newChild = this.onAddNewField(targetItem, dropItem, $insertAt);
                }
                else {
                    if ($insertAt) {
                        newChild = this.addNewSection(this.targetPage.layoutContent, null, 0, dropItem.$category);
                        newChild.layoutParent.extractItem(newChild);
                    }
                    newChild = this.addNewItem(targetItem, dropItem.$authoringLevel, null, newChild, $insertAt, dropItem.$category);
                }
            }
            else {
                newChild = this.addNewItem(targetItem, dropItem.$authoringLevel, dropItem.$bind, dropItem, $insertAt);
            }
        }
        this.endTargetPageUpdate(newChild, true);
    },
    switchTabPanel: function(item, $layoutType){
        if (item.layoutParent.items.length == 1) {
            this.newConvert(item.layoutParent, {
                $layoutType: $layoutType
            });
        }
        else {
            var layoutRoot = item.layoutParent;
            if ($layoutType == "stack") {
                this.newConvert(layoutRoot, {
                    $layoutType: "stack"
                });
                var children = this.getItemsFromLayout(layoutRoot);
                var layoutTab;
                for (var ii = 0, jj = children.length; ii < jj; ii++) {
                    if (children[ii] == item) {
                        layoutRoot.loadChildItems([children[ii]]);
                        layoutTab = null;
                    }
                    else {
                        if (!children[ii].isSpaceBox) {
                            if (!layoutTab) {
                                layoutTab = layoutRoot.loadChildItem(null, {
                                    $layoutType: "tabs"
                                });
                            }
                            layoutTab.loadChildItems([children[ii]]);
                        }
                        else {
                            // layoutRoot.removeItem(children[ii], true);
                        }
                    }
                }
            }
            else {
                var layoutTab = layoutRoot.wrapIntack([item], layoutRoot.getItemIndex(item.$item));
                this.newConvert(layoutTab, {
                    $layoutType: "tabs"
                });
            }
        }
    },
    newConvert: function(layout, $newLayout){
        var $oldLayoutType = layout.$layout.$layoutType;
        var $oldWidths = layout.$layout.$widths;
        var isTabLayout = $newLayout.$layoutType == "tabs";
        if (layout._tabs && !isTabLayout) {
            document.site.removeDomChild(layout._tabs);
            delete layout._tabs;
        }
        //replace layout properties by new ones
        delete layout.$layout.$autoSize;
        delete layout.$layout.$widths;
        if ($newLayout.$autoSize != undefined) {
            layout.$layout.$autoSize = $newLayout.$autoSize;
        }
        if ($newLayout.$widths != undefined) {
            layout.$layout.$widths = $newLayout.$widths;
        }
        layout.$layout.$layoutType = $newLayout.$layoutType;
        layout.ensureLayoutMode();
        
        if ($oldLayoutType == "row") {
            var oldWidths = $oldWidths.split(",");
            var newWidths = (layout.$layout.$widths || "100").split(",");
            if (oldWidths.length > newWidths.length) {
                var targetCell = layout.items[newWidths.length - 1]
                var targetLayout = targetCell;
                var outofItems = layout.items.slice(newWidths.length);
                for (var ii = 0, jj = outofItems.length; ii < jj; ii++) {
                    var item = outofItems[ii];
                    if (item.isSpaceBox) {
                        layout.removeItem(item, true);
                    }
                    else {
                        if (targetCell.isSpaceBox) {
                            layout.removeItem(targetCell, true);
                        }
                        if (targetCell.isLayout) {
                            if (targetCell.$layout.$layoutType == "stack") {
                                if (item.isLayout && item.$layout.$layoutType == "stack") {
                                    item.moveItems(targetCell);
                                    layout.removeItem(item, true);
                                }
                                else {
                                    targetCell.loadChildItems([item]);
                                }
                            }
                            else {
                                targetLayout = layout.wrapIntack([targetCell], layout.getItemIndex(targetCell.$item));
                                targetLayout.loadChildItems([item]);
                            }
                        }
                        else {
                            if (targetCell.isDisposed) {
                                //previous was removed spaceBox
                                layout.loadChildItems([item]);
                            }
                            else {
                                targetLayout = layout.wrapIntack([targetCell], layout.getItemIndex(targetCell.$item));
                                targetLayout.loadChildItems([item]);
                            }
                        }
                    }
                }
            }
        }
        /* else {
         if (layout.items.length > 1) {
         layout.wrapIntack(layout.items.slice(0));
         }
         }*/
        if (isTabLayout) {
            var children = this.getItemsFromLayout(layout);
            /* for (var ii = 0, jj = children.length; ii < jj; ii++) {
             // if (!children[ii].isSpaceBox) {
             layout.loadChildItem(children[ii], undefined, ii);
             // }
             
             }*/
            layout.loadChildItems(children);
            var toClean = layout.items.slice(0);
            for (var ii = 0, jj = toClean.length; ii < jj; ii++) {
                var item = toClean[ii];
                if (item.isSpaceBox || !item.isSection) {
                    layout.removeItem(item, true);
                }
            }
        }
        else {
            if (layout.$layout.$layoutType == "stack") {
                if (!layout.isSpaceBox && layout.layoutParent && layout.items.length == 1) {
                    var layoutParent = layout.layoutParent;
                    layout.moveItems(layoutParent, null, layoutParent.getItemIndex(layout.$item));
                    layoutParent.removeItem(layout, true);
                    layout = layoutParent;
                }
            }
        }
        for (var ii = 0, jj = layout.items.length; ii < jj; ii++) {
            var item = layout.items[ii];
            item.isTabLayout = isTabLayout;
            item.ensureLayoutMode();
            if (!item.isLayout && !isTabLayout) {
                if (item.isSection && !item.tabTitle && !item.loaded && item.$item.$opened !== false) {
                    item.openBox(true);
                }
            }
            else {
                if ($oldLayoutType == "tabs" && !isTabLayout) {
                    for (var mm = 0, kk = item.items.length; mm < kk; mm++) {
                        var child = item.items[mm];
                        if (!child.isLayout && child.isSection && !child.tabTitle && !child.loaded && child.$item.$opened !== false) {
                            child.openBox(true);
                        }
                    }
                }
            }
        }
        return layout;
    },
    excludeItem: function(removedItem){
        if (removedItem.isMenuItem && this.targetPage.isDashBoard) {
            this.targetPage.garbageLinks[removedItem.$bind] = 1;
        }
        var layoutParent = removedItem.layoutParent;
        var excludedFields = (removedItem.$authoringLevel == "field") ? [removedItem] : this._extractFields(removedItem.layoutContent);
        if (excludedFields) {
            for (var ii = 0, jj = excludedFields.length; ii < jj; ii++) {
                this.palette.toolsBar.onExcludeField(excludedFields[ii].$item.$bind, false);
            }
        }
        var index = layoutParent.getItemIndex(removedItem, true);
        layoutParent.removeItem(removedItem, true);
        var newItem = layoutParent.items[index];
        if (!newItem) {
            newItem = layoutParent.items[--index];
        }
        newItem = newItem || layoutParent.boxParent;
        switch (layoutParent.$layout.$layoutType) {
            case "row":
                layoutParent.addSpaceBox(index);
                break;
            case "tabs":
                if (layoutParent.items.length == 0) {
                    this.newConvert(layoutParent, {
                        $layoutType: "stack",
                        $width: "100"
                    });
                }
                break;
        }
        if (this.palette.toolsBar && this.palette.toolsBar.treeContent) {
            this.palette.toolsBar.treeContent.toggle(true);
        }
        return newItem;
    },
    dispose: function(){
        if (this.$viewType != "layout") {
            this.toggleViewType("layout");
        }
        if (this.rowResizer) {
            this.rowResizer.dispose();
            this.rowResizer = null;
        }
        if (this.targetPage) {
            if (this.targetPage.dialogWrapper) {
                this._toggleAuthoringDialog(false);
            }
            this.targetPage.isOnAuthoring = false;
            this.isOnPreviewMode = false;
            this.toggleItemAuthoring(this.targetPage, false);
            delete this.targetPage.authoringPage;
            this.diagnosePage = this.targetPage = null;
        }
        if (this.paletteSlot) {
            document.site.removeDomChild(this.paletteSlot);
            this.paletteSlot = null;
        }
        this._overItem = this._oveItems = this.targetPage = this.awItem = this.awLayout = null;
        if (document.site.$$layoutSlot) {
            document.site.$$layoutSlot.undelegate(".author");
        }
        this._ddTabEnter = this.page = this.paletteSlot = this._paletteBar = null;
        RawPage.prototype.dispose.call(this);
    }
});
