// syracuse-ui/lib/authoring/Copy of landingPageDesigner.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var PageDesigner = require("syracuse-ui/lib/authoring/pageDesigner").PageDesigner;
var UISwitchTool = require("syracuse-ui/lib/authoring/tools/uiSwitchTool").UISwitchTool;

function LandingPageDesigner() {}

exports.LandingPageDesigner = helpers.defineClass(LandingPageDesigner, PageDesigner, {
	initializeDesigner: function() {
		//important override default desktop page 
	},
	drawBox: function() {
		this.layoutSlot = document.createElement("aside");
		this.layoutSlot.className = "s-aw-lp-designer";
		this.designedItem = this.designedArticle;
		this.designedArticle.header.appendChild(this.layoutSlot);
		this.isVisible = false;
		this.toggleDesigner(false);
		this.$$item = $(this.layoutSlot);
		this.addClosePicker(this.layoutSlot);
		this.sectionModel = {
			body: this.layoutSlot
		};
		(this.uiSwitchTool = new UISwitchTool()).load(this);

		this.loadTemplateTools();
		document.site.ddManager.togglePageObserver(this, true);
		this.toggleUIDesign(this.designedArticle, true);
		this.bindEvents(true);
		document.site.resize();
	},
	onMenuClick: function(menuItem) {
		if (menuItem.$bind == "$select") {

			return false;
		} else {
			PageDesigner.prototype.onMenuClick.call(this, menuItem);
			return true;
		}
	},
	onClickItem: function(item, target, article) {
		PageDesigner.prototype.onClickItem.call(this, item, target, article);
		return true;
	},
	onTabClik: function(tabOpened) {
		if (this.isVisible) {
			this.templatesTool.addLandingPageModels(this.sectionModel, tabOpened || this.designedItem.layoutContent.tabOpened);
		}
	},
	selectItem: function() {

	},
	toggleDesigner: function(show) {
		this.isVisible = show === undefined ? !this.isVisible : show;
		document.site.toggleClass(this.designedArticle.designerOpenerLink, "s-close", this.isVisible);
		this.layoutSlot.style.display = this.isVisible ? "" : "none";
		this.onTabClik();
		document.site.resize();
	},
	onTemplateChoiceClick: function(picker, designedItem, awLayout) {
		designedItem = designedItem.layoutContent.tabOpened;
		if (picker.syraFieldsCount) {
			var funcName;
			var templates = this.templatesTool.countTemplates[picker.syraFieldsCount];
			if (templates) {
				funcName = picker.syraFieldsCount;
			} else {
				templates = this.templatesTool.defaultTemplates;
				funcName = picker.syraTemplate;
			}
			var fields = designedItem.layoutContent.getFields();
			designedItem.layoutContent.extractItems(fields);
			this.templatesTool["build_" + funcName](designedItem, picker.syraTemplate, fields);
			return designedItem;
		}
		return null;
	},
	dispose: function() {
		this.sectionModel = null;
		PageDesigner.prototype.dispose.call(this);
	}
});
