// syracuse-ui/lib/authoring/designStore.js
"use strict";
var _advancedStateHelper = require('syracuse-ui/lib/field/helpers/advancedStateHelper');

exports.addPageOpener = function() {
	var btn = syra_site.pageDesignerOpener = syra_menus.addIconButton(syra_local.userProfile_siteAuthoringOpen, "s-page-designer-opener", "onOpenerClick");
	btn.syraSiteObserver = "designStore";
	btn.style.visibility = "hidden";
	syra_site.headerTop.appendChild(btn);
};
exports.showPageOpener = function(show) {
	if (syra_site.pageDesignerOpener) {
		syra_site.pageDesignerOpener.style.visibility = show ? "" : "hidden";
	}
};
exports.onOpenerClick = function() {
	var page;
	syra_site.topPanel.toggle(false);
	if (syra_site.mainPage && syra_site.mainPage.isFusionPage) {
		page = syra_site.fusionGateway.activatedBook.selectedSheet;
	} else {
		page = syra_site.dialogManager.getTopDialogPage();
		if (page) {
			page = page._content;
		}
	}
	page = page || syra_site.mainPage;
	if (page) {
		if (syra_site.currentQuickEdit &&
			syra_site.currentQuickEdit.needCloseConfirm(page, function() {
				exports.onOpenerClick();
			})) {
			return;
		}
		syra_site.designStore.design(page, true);
	}

};

var _isCssLoaded;

function _loadCss() {
	if (!_isCssLoaded) {
		_isCssLoaded = true;
		syra_site.dom.loadStyleSheet("authoring", ["authoring.css"]);
	}
}


var _designer = {
	page: function(page, open) {
		if (open) {
			var path = page.isLandingPage ? "syracuse-ui/lib/authoring/landingPageDesigner" : "syracuse-ui/lib/authoring/pageDesigner";
			require.async(path, function(err, module) {
				_advancedStateHelper.toggleAllFields(page, true);
				page.designer = new module.Designer();
				page.designer.openDesigner(page);
				page.resizeArticle(true);
			});
		} else {
			if (page.designer) {
				syra_site.showDiagnoses({
					$diagnoses: null
				}, page.page);
				page.designer.dispose();
				_advancedStateHelper.toggleAllFields(page);
				page.resizeArticle(true);
			}
			page.designer = null;
		}
	},
	list: function(list, open) {
		if (open) {
			require.async('syracuse-ui/lib/authoring/listDesigner', function(err, module) {
				if (list.graphDecorator) {
					list.graphDecorator.ensureGraph(true);
				}
				list.ensureGrapDecorator();
				list.designer = new module.Designer();
				list.designer.openDesigner(list);
				list.onAfterSwitchDesign(open);
			});
		} else {
			list.designer && list.designer.dispose();
			list.designer = list.$designing = null;
			list.onAfterSwitchDesign(open);
		}
	},
	htmlField: function(field, open) {
		if (open) {
			require.async('syracuse-ui/lib/authoring/htmlVignetteDesigner', function(err, module) {
				syra_site.dom.toggleClass(field.btns.items.design, "s-close", open);
				field.designer = new module.Designer();
				field.designer.openDesigner(field);
			});
		} else {
			if (field.designer) {
				syra_site.dom.toggleClass(field.btns.items.design, "s-close", open);
				this.showDiagnoses({
					$diagnoses: null
				}, field.page);
				field.designer.dispose();
			}
			field.designer = null;
		}
	}
};


exports.design = function(item, open) {
	var designer = _designer[item.$designerClass];
	if (designer) {
		open && _loadCss();
		designer(item, open);
	}
};


exports.designCardItem = function(listDesigner, cardItem, open) {
	if (open) {
		open && _loadCss();
		require.async('syracuse-ui/lib/authoring/landingPageDesigner', function(err, module) {
			listDesigner.cardDesigner = cardItem.designer = new module.Designer();
			cardItem.designer.openDesigner(cardItem);
			cardItem.designer.resizeArticle();
		});
	} else {
		cardItem.designer && cardItem.designer.dispose();
		cardItem.designer = null;
		syra_site.resize();
	}
};
