// syracuse-ui/lib/authoring/designer.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var RawPage = require("syracuse-ui/lib/page/rawPage").RawPage;
var _historyTool = require("syracuse-ui/lib/authoring/tools/historyTool");
var _templatesTool = require("syracuse-ui/lib/authoring/tools/templatesTool");

function _clearEmptySection(section) {
	if (section && section.page != section) {
		var boxParent = section.boxParent;
		var layoutParent = section.layoutParent;
		layoutParent && layoutParent.removeItem(section, true);
		if (boxParent) {
			if (!(boxParent.boxChildItems && boxParent.boxChildItems.length > 0) && !(boxParent.childrenSection && boxParent.childrenSection.length > 0)) {
				_clearEmptySection(boxParent);
			}
		}
	}
}


function _clearUnknowFields(page) {
	if (page.noProtoFields) {
		for (var ii = 0, jj = page.noProtoFields.length; ii < jj; ii++) {
			var field = page.noProtoFields[ii];
			var boxParent = field.boxParent;
			if (field.layoutParent) {
				field.layoutParent.removeItem(field, true);
			}
			if (boxParent) {
				if (!(boxParent.boxChildItems && boxParent.boxChildItems.length > 0) && !(boxParent.childrenSection && boxParent.childrenSection.length > 0)) {
					_clearEmptySection(boxParent);
				}
			}
		}
		delete page.noProtoFields;
	}
}


function Designer() {}

exports.Designer = helpers.defineClass(Designer, RawPage, {
	addCloseButton: function() {
		return syra_menus.addIconButton(syra_local.aw_close, "s-aw-designer-close", "onCloseDesignerClick");
	},
	loadHistoryTool: function() {
		_historyTool.load(this);
	},
	endArticleUpdate: function(selectedItem, isStructureUpdated) {
		syra_site.ensureArticleVisibility(this.designedArticle.page);
		this.history && this.history.updateSteps();
		this.saveDesign();
	},
	showTargetPageDiagnoses: function(message, options) {
		syra_site.showDiagnoses(message, this.designedArticle, options);
		if (syra_site.mobileGateway) {
			syra_site.mobileGateway.onShowTargetPageDiagnoses(message, options);
		}
	},
	_toggleDesignDialog: function(page, begin) {
		var dlg = page && page.dialogWrapper;
		if (dlg) {
			dlg.btn_close.style.display = begin ? "none" : "";
			dlg.overlay.className = begin ? "s-aw-overlay" : "s-overlay";
			!begin && dlg.resizeDialog();
		}
	},

	openDesigner: function(designedArticle) {
		this.isSiteRegisterDisabled = true;
		this.designedArticle = designedArticle;
		this.designedArticle.designer = this;
		this.designedArticle.isDesigned = true;
		this.$prototype = {
			$properties: {},
			$links: {}
		};
		this.$item = {};
		if (this.designedArticle.page == this.designedArticle) {
			_clearUnknowFields(this.diagnosePage = this.designedArticle);
			this.showTargetPageDiagnoses({
				$diagnoses: null
			});
		}
		if (this.designedArticle.page.dialogWrapper) {
			this._toggleDesignDialog(this.designedArticle.page, true);
		}
		if (designedArticle.$prototype && designedArticle.$prototype.$article) {
			//imortant save article for page
			designedArticle.$prototype.$article = helpers.object.clone(designedArticle.$prototype.$article, true);
		}
		this.loadBox(designedArticle);
		this.isPageLoaded = true;
		if (!this.designedArticle.page.isVignettePage) {
			syra_site.ensureArticleVisibility(this);
		}
	},
	notifyDataChange: function(field, value) {
		var metaData = {};
		metaData[field.$item.$bind] = value;
		this.designedItem.applyDesignMetaData(metaData, true);
		this.saveDesign();
		this.history && this.history.updateSteps();
		this.applyChange(metaData);
	},
	loadTemplateTools: function() {
		_templatesTool.load(this);
	},
	onCloseDesignerClick: function() {
		var self = this;
		setTimeout(function() {
			self.closeDesigner();
		}, 100);
	},
	findItem: function(domItem, findLayout) {
		if (findLayout || domItem.syraDesignedLayout) {
			while (domItem && !domItem.syraDesignedLayout) {
				domItem = domItem.parentNode;
			}
			return (domItem && domItem.syraDesignedLayout) || null;
		} else {
			while (domItem && !domItem.syraDesignedItem) {
				domItem = domItem.parentNode;
			}
			if (domItem) {
				return domItem.syraDesignedItem || syra_store.findItem(domItem) || null;
			}
		}
		return null;
	},
	dispose: function() {
		var page;
		if (this.designedArticle) {
			page = this.designedArticle.page;
			this.designedArticle.isDesigned = false;
			delete this.designedArticle.designer;
		}
		RawPage.prototype.dispose.call(this);
		this._toggleDesignDialog(page, false);
	}
});
