// syracuse-ui/lib/authoring/historyHelper.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var patternFactory = require("syracuse-ui/lib/authoring/patterns/patternFactory");

function _refreshStepLinks(authorPage, awContext){
    var metadata;
    if (awContext.updatedSteps.length > 1) {
        metadata = {
            $links: {
                $previousLayout: {
                    $title: authorPage.localize.aw_previousLayoutOn.replace("{0}", awContext.currentUpdtateStep).replace("{1}", awContext.updatedSteps.length),
                    $isDisabled: awContext.currentUpdtateStep == 0
                },
                $undoAllLayout: {
                    $isDisabled: false
                }
            }
        };
        if (awContext.currentUpdtateStep == awContext.updatedSteps.length - 1) {
            metadata.$links.$nextLayout = {
                $title: authorPage.localize.aw_nextLayout,
                $isDisabled: true
            };
        }
        else {
            metadata.$links.$nextLayout = {
                $title: authorPage.localize.aw_nextLayoutOn.replace("{0}", awContext.currentUpdtateStep + 1).replace("{1}", awContext.updatedSteps.length),
                $isDisabled: false
            };
        }
    }
    else {
        metadata = {
            $links: {
                $previousLayout: {
                    $title: authorPage.localize.aw_previousLayout,
                    $isDisabled: true
                },
                $nextLayout: {
                    $title: authorPage.localize.aw_nextLayout,
                    $isDisabled: true
                },
                $undoAllLayout: {
                    $isDisabled: true
                }
            }
        };
    }
    authorPage.palette.applyChange(metadata);
}

function _onEndChangeStep(authorPage, awContext, menuItem, $item){
    var $bind = authorPage.awItem.$item.$bind;
    authorPage.targetPage.reloadLayout(helpers.object.clone($item || awContext.updatedSteps[awContext.currentUpdtateStep], true));
    _refreshStepLinks(authorPage, awContext);
    var selectItem;
    if ($bind) {
        var bounds = authorPage.targetPage.boundFields[$bind];
        selectItem = bounds ? bounds[0] : null;
    }
    authorPage.palette.toolsBar.onEndChangeStep();
    authorPage.selectItem(selectItem || authorPage.targetPage, true);
}


exports.notifySteps = function(authorPage, awContext, resetUpdates){
    if (!resetUpdates && authorPage.isUpdated) {
        if (awContext.currentUpdtateStep != awContext.updatedSteps.length - 1) {
            awContext.updatedSteps.splice(awContext.currentUpdtateStep + 1);
            //this.notifySteps(true);
        }
        awContext.updatedSteps.push(helpers.object.clone(authorPage.targetPage.$item, true));
        awContext.currentUpdtateStep = awContext.updatedSteps.length - 1;
    }
    else {
        (awContext.updatedSteps = []).push(authorPage.$sourceItem);
        awContext.currentUpdtateStep = 0;
    }
    _refreshStepLinks(authorPage, awContext);
};


exports.onChangeSep = function(authorPage, awContext, menuItem){
    var $item;
    switch (menuItem.$item.$bind) {
        case "$previousLayout":
            awContext.currentUpdtateStep--;
            _onEndChangeStep(authorPage, awContext, menuItem, $item);
            break;
        case "$nextLayout":
            awContext.currentUpdtateStep++;
            _onEndChangeStep(authorPage, awContext, menuItem, $item);
            break;
        case "$undoAllLayout":
            document.site.showMessage({
                $title: authorPage.localize.aw_updateMessageTitle,
                $message: authorPage.localize.aw_confirmUndoAll,
                $type: "question",
                $buttons: "yesno",
                callback: function(response){
                    if (response.$id == "yes") {
                        awContext.currentUpdtateStep = 0;
                        exports.notifySteps(authorPage, awContext, true);
                        _onEndChangeStep(authorPage, awContext, menuItem, $item);
                    }
                }
            });
            break;
        case "$defaultLayout":
            $item = authorPage.$defaultItem;
            authorPage.notifyUpdate(true);
            _onEndChangeStep(authorPage, awContext, menuItem, $item);
            break;
        default:
            if (menuItem.$item.$bind.indexOf("pattern") == 0) {
                if ((menuItem.$item.$bind != "pattern-headerTabs") && (authorPage.awItem.$authoringLevel == "section" || authorPage.awItem.$authoringLevel == "block")) {
                    var pattern = patternFactory.patterns[menuItem.$item.$bind];
                    if (pattern.isBoxOnly) {
                        var children = authorPage.getItemsFromLayout(authorPage.awItem.layoutContent);
                        if (children.length && children[0].$authoringLevel == "field") {
                            return;
                        }
                    }
                    var $result = patternFactory.build(menuItem.$item.$bind, authorPage.targetPage.$prototype, authorPage.awItem.$item);
                    if (!$result && !$result.$layout) {
                        return;
                    }
                    authorPage.awItem.$item.$layout = $result.$layout;
                    $item = authorPage.targetPage.$item;
                }
                else {
                    $item = patternFactory.build(menuItem.$item.$bind, authorPage.targetPage.$prototype, authorPage.$sourceItem);
                }
                authorPage.notifyUpdate(true);
            }
            _onEndChangeStep(authorPage, awContext, menuItem, $item);
            break;
    }
};
