// syracuse-ui/lib/authoring/historyHelper.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var patternFactory = require("syracuse-ui/lib/authoring/patterns/patternFactory");

function _refreshStepLinks(authorPage){
    var metadata;
    if (authorPage.updatedSteps.length > 1) {
        metadata = {
            $links: {
                $previousLayout: {
                    $title: authorPage.localize.aw_previousLayoutOn.replace("{0}", authorPage.currentUpdtateStep).replace("{1}", authorPage.updatedSteps.length),
                    $isDisabled: authorPage.currentUpdtateStep == 0
                },
                $undoAllLayout: {
                    $isDisabled: false
                }
            }
        };
        if (authorPage.currentUpdtateStep == authorPage.updatedSteps.length - 1) {
            metadata.$links.$nextLayout = {
                $title: authorPage.localize.aw_nextLayout,
                $isDisabled: true
            };
        }
        else {
            metadata.$links.$nextLayout = {
                $title: authorPage.localize.aw_nextLayoutOn.replace("{0}", authorPage.currentUpdtateStep + 1).replace("{1}", authorPage.updatedSteps.length),
                $isDisabled: false
            };
        }
    }
    else {
        metadata = {
            $links: {
                $previousLayout: {
                    $title: authorPage.localize.aw_previousLayout,
                    $isDisabled: true
                },
                $nextLayout: {
                    $title: authorPage.localize.aw_nextLayout,
                    $isDisabled: true
                },
                $undoAllLayout: {
                    $isDisabled: true
                }
            }
        };
    }
    authorPage.topBar.applyChange(metadata);
}

function _onEndChangeStep(pageContext, menuItem, $item){
    var $bind = pageContext.awItem.$item.$bind;
    var authorPage = pageContext.authorPage;
    pageContext.awArticle.reloadLayout(helpers.object.clone($item || authorPage.updatedSteps[authorPage.currentUpdtateStep], true));
    _refreshStepLinks(authorPage);
    var selectItem;
    if ($bind) {
        var bounds = pageContext.awArticle.boundFields[$bind];
        selectItem = bounds ? bounds[0] : null;
    }
    pageContext.toolsBar.onEndChangeStep();
    authorPage.selectItem(selectItem || pageContext.awArticle, true);
}

exports.notifySteps = function(awContext, resetUpdates){
    var pageContext = awContext.authorPage.getPageContext();
    var authorPage = pageContext.authorPage;
    if (!resetUpdates && authorPage.isUpdated) {
        if (authorPage.currentUpdtateStep != authorPage.updatedSteps.length - 1) {
            authorPage.updatedSteps.splice(authorPage.currentUpdtateStep + 1);
            //this.notifySteps(true);
        }
        authorPage.updatedSteps.push(helpers.object.clone(pageContext.awArticle.$item, true));
        authorPage.currentUpdtateStep = authorPage.updatedSteps.length - 1;
    }
    else {
        (authorPage.updatedSteps = []).push(pageContext.$sourceItem);
        authorPage.currentUpdtateStep = 0;
    }
    _refreshStepLinks(authorPage);
};

exports.onChangeSep = function(pageContext, menuItem){
    var $item;
    var authorPage = pageContext.authorPage;
    switch (menuItem.$item.$bind) {
        case "$previousLayout":
            authorPage.currentUpdtateStep--;
            _onEndChangeStep(pageContext, menuItem, $item);
            break;
        case "$nextLayout":
            authorPage.currentUpdtateStep++;
            _onEndChangeStep(pageContext, menuItem, $item);
            break;
        case "$undoAllLayout":
            document.site.showMessage({
                $title: authorPage.localize.aw_updateMessageTitle,
                $message: authorPage.localize.aw_confirmUndoAll,
                $type: "question",
                $buttons: "yesno",
                callback: function(response){
                    if (response.$id == "yes") {
                        authorPage.currentUpdtateStep = 0;
                        exports.notifySteps(pageContext, true);
                        _onEndChangeStep(pageContext, menuItem, $item);
                    }
                }
            });
            break;
        case "$defaultLayout":
            $item = pageContext.$defaultItem;
            authorPage.notifyUpdate(true);
            _onEndChangeStep(pageContext, menuItem, $item);
            break;
        default:
            if (menuItem.$item.$bind.indexOf("pattern") == 0) {
                if ((menuItem.$item.$bind != "pattern-headerTabs") && (pageContext.awItem.$authoringLevel == "section" || pageContext.awItem.$authoringLevel == "block")) {
                    var pattern = patternFactory.patterns[menuItem.$item.$bind];
                    if (pattern.isBoxOnly) {
                        var children = pageContext.drawHelper.getItemsFromLayout(pageContext.awItem.layoutContent);
                        if (children.length && children[0].$authoringLevel == "field") {
                            return;
                        }
                    }
                    var $result = patternFactory.build(menuItem.$item.$bind, pageContext.$prototype, pageContext.awItem.$item);
                    if (!$result && !$result.$layout) {
                        return;
                    }
                    pageContext.awItem.$item.$layout = $result.$layout;
                    $item = pageContext.awArticle.$item;
                }
                else {
                    $item = patternFactory.build(menuItem.$item.$bind, pageContext.$prototype, pageContext.$sourceItem);
                }
                authorPage.notifyUpdate(true);
            }
            _onEndChangeStep(pageContext, menuItem, $item);
            break;
    }
};
