// syracuse-ui/lib/authoring/htmlVignetteDesigner.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Designer = require("syracuse-ui/lib/authoring/designer").Designer;

function HTMLVignetteDesigner() {}

exports.HTMLVignetteDesigner = helpers.defineClass(HTMLVignetteDesigner, Designer, {
	loadBox: function(designedPage) {
		this.$skin = "s-vignette-designer";
		this.$prototype = {
			$properties: {
				$title: {
					$type: "application/x-string",
					$title: this.localize.htmlVignetteTitle,
					$displayLength: 200
				},
				$url: {
					$type: "application/x-string",
					$title: this.localize.htmlVignetteUrl,
					$displayLength: 200
				},
				$height: {
					$type: "application/x-string",
					$title: this.localize.htmlVignetteHeight,
					$displayLength: 8
				}
			}
		};
		this.$item = {
			$category: "section",
			$layout: {
				$items: [{
					$bind: "$url",
					$isEditMode: true,
					$isTopLabelAlignment: false,
					$css: "s-aw-prop"
				}, {
					$layoutType: "row",
					$autoSize: true,
					$items: [{
						$bind: "$title",
						$isEditMode: true,
						$isTopLabelAlignment: false,
						$css: "s-aw-prop"
					}, {
						$bind: "$height",
						$isEditMode: true,
						$isTopLabelAlignment: false,
						$css: "s-aw-prop"
					}]
				}]
			}
		};
		Designer.prototype.loadBox.call(this);
		this.applyChange({
			$title: this.awItem.$item.$title || this.awItem.$field.$title,
			$url: this.awItem.$item.$url || (this.awItem.$field.$location ? this.awItem.$field.$location.$url : null),
		});
	},
	drawBox: function() {
		this.layoutSlot = document.createElement("aside");
		this.layoutSlot.className = "s-aw-vignette-designer";
		this.awArticle.body.insertBefore(this.layoutSlot, this.awArticle.body.firstChild);
		this.layoutSlot.style.display = "";
		this.addClosePicker(this.layoutSlot);
		this.$$item = $(this.body =this.layoutSlot);
		this.renderLayoutContent();
		this.awItem = this.awArticle;		
		document.site.resize();
	},
	onWindowResize: function() {

	},
	dispose: function() {
		document.site.removeDomChild(this.layoutSlot);
		Designer.prototype.dispose.call(this);
	}
});
