// syracuse-ui/lib/authoring/htmlVignetteDesigner.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Designer = require("syracuse-ui/lib/authoring/designer").Designer;

function HTMLVignetteDesigner() {}

exports.HTMLVignetteDesigner = helpers.defineClass(HTMLVignetteDesigner, Designer, {
	saveDesign: function() {
		if (this.designedArticle.page.designer) {
			this.designedArticle.page.designer.endArticleUpdate();
		} else {
			this.designedArticle.page.savePageDesign();
		}
	},
	loadBox: function(designedPage) {
		this.$skin = "s-vignette-designer";
		this.$prototype = {
			$properties: {
				$title: {
					$type: "application/x-string",
					$title: this.localize.htmlVignetteTitle,
					$displayLength: 200
				},
				$url: {
					$type: "application/x-string",
					$isHidden: true,
					$title: this.localize.htmlVignetteUrl,
					$displayLength: 200
				},
				$height: {
					$type: "application/x-string",
					$title: this.localize.htmlVignetteHeight,
					$displayLength: 8
				}
			}
		};
		this.$item = {
			$layout: {
				$items: [{
					$category: "section",
					$skin: "s-aw-h2",
					$layout: {
						$items: [{
							$bind: "$url",
							$isEditMode: true,
							$isTopLabelAlignment: false,
							$css: "s-aw-prop"
						}, {
							$layoutType: "row",
							$autoSize: true,
							$items: [{
								$bind: "$title",
								$isEditMode: true,
								$isTopLabelAlignment: false,
								$css: "s-aw-prop"
							}, {
								$bind: "$height",
								$isEditMode: true,
								$isTopLabelAlignment: false,
								$css: "s-aw-prop"
							}]
						}]
					}
				}]
			}
		};
		Designer.prototype.loadBox.call(this);
		this.applyChange({
			$title: this.designedItem.$item.$title || this.designedItem.$field.$title,
			$url: this.designedItem.$item.$url || (this.designedItem.$field.$location ? this.designedItem.$field.$location.$url : null),
			$height: this.designedItem.$item.$height || this.designedItem.$field.$height
		});
	},
	drawBox: function() {
		this.layoutSlot = document.createElement("aside");
		this.layoutSlot.className = "s-aw-vignette-designer";
		this.designedArticle.body.insertBefore(this.layoutSlot, this.designedArticle.body.firstChild);
		this.layoutSlot.style.display = "";
		this.addClosePicker(this.layoutSlot);
		this.$$item = $(this.body = this.layoutSlot);
		this.loadHistoryTool();
		this.renderLayoutContent();
		this.designedItem = this.designedArticle;
		document.site.resize();
	},
	closeDesigner: function() {
		this.designedArticle.designArticle(false);
	},
	dispose: function() {
		document.site.removeDomChild(this.layoutSlot);
		Designer.prototype.dispose.call(this);
	}
});
