// syracuse-ui/lib/authoring/htmlVignetteDesigner.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Designer = require("syracuse-ui/lib/authoring/designer").Designer;

function HTMLVignetteDesigner() {}

exports.HTMLVignetteDesigner = helpers.defineClass(HTMLVignetteDesigner, Designer, {
	loadBox: function(designedPage) {
		this.$skin = "s-vignette-designer";
		this.$prototype = {
			$properties: {
				$title: {
					$type: "application/x-string",
					$title: this.localize.htmlV_location_title
				},
				$url: {
					$type: "application/x-string",
					$title: this.localize.htmlV_location_url
				},
				$height: {
					$type: "application/x-string",
					$title: this.localize.htmlV_height
				}
			}
		};
		this.$item = {
			$category: "section",
			$layout: {
				$items: [{
					$bind: "$url",
					$isEditMode: true,
					$skin: "s-aw-field"
				}, {
					$layoutType: "row",
					$autoSize: true,
					$items: [{
						$bind: "$title",
						$isEditMode: true,
						$skin: "s-aw-field"
					}, {
						$bind: "$height",
						$isEditMode: true,
						$skin: "s-aw-field"
					}]
				}]
			}
		};
		Designer.prototype.loadBox.call(this);
		this.applyChange({
			$title: this.awItem.$item.$title || this.awItem.$field.$title,
			$url: this.awItem.$item.$url || (this.awItem.$field.$location ? this.awItem.$field.$location.$url : null),
		});
	},
	drawBox: function() {
		this.layoutSlot = document.createElement("aside");
		this.layoutSlot.className = "s-aw-page";
		this.awArticle.body.insertBefore(this.layoutSlot, this.awArticle.body.firstChild);
		this.layoutSlot.style.display = "";
		this.$$item = this.$$body = $(this.layoutSlot);
		this.renderLayoutContent();

		this.awItem = this.awArticle;

		this.loadDefaultTools();
		this.workinCopySlot.appendChild(this.historySlot);
		this.layoutSlot.appendChild(this.workinCopySlot);

		document.site.resize();
	},
	onWindowResize: function() {

	},
	dispose: function() {
		document.site.removeDomChild(this.layoutSlot);
		Designer.prototype.dispose.call(this);
	}
});
