// syracuse-ui/lib/authoring/insertTree.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');


function InsertTree() {}

exports.InsertTree = helpers.defineClass(InsertTree, null, {
	load: function(awPalette, awArticle, slot) {
		this.awPalette = awPalette;
		slot.style.display = "none";
		document.site.emptyDom(slot);
		this._nodes = [];
		this._addNode(awArticle.$item, {
			node: slot,
			isRoot: true
		});
		(this.slot = slot).style.display = "";
	},
	findField: function($bind) {
		for (var ii = 0, jj = this._nodes.length; ii < jj; ii++) {
			if (this._nodes[ii].$item.$bind == target.$bind) {
				return this._nodes[ii];
			}
		}
	},
	unload: function() {
		this.slot.style.display = "none";
		document.site.emptyDom(this.slot);
		this.dispose();
	},
	dispose: function() {
		if (this._nodes) {
			for (var ii = 0, jj = this._nodes.length; ii < jj; ii++) {
				var node = this._nodes[ii];
				if (node) {
					node.item = node.node = node.children = node.data = null;
					node.opener = node.title = null;
				}
			}
		}
		this.selectedNode = this._nodes = this.slot = this.awPalette = null;
	},
	onNodeEvent: function(event) {
		var authorPage = document.site.authorPage;
		delete document.site.requestedDDAuthoringItem;
		event.stopPropagation();
		var node = this.findNode(event);
		if (node) {
			if (event.target.className.indexOf("s-aw-tree-item-opener-children") >= 0) {
				var open = !(node.opener.className.indexOf("s-open") < 0);
				node.children.style.display = open ? "" : "none";
				document.site.toggleClass(node.opener, "s-open", !open);
			} else {
				//event hidden  if card
				switch (event.type) {
					case "mouseenter":
						authorPage.toggleOverItem(node.item, true);
						this.awPalette.targetPage.scrollToItem(node.item);
						break;
					case "mouseleave":
						authorPage.toggleOverItem(node.item, false);
						break;
					case "click":
						if (node.item == this.awPalette.awArticle) {
							this.awPalette.targetPage.scrollToItem(node.item);
						}
						this.selectNode(node.item, true);
						this._isSelecteDisabled = true;
						authorPage.selectItem(node.item, true);
						this._isSelecteDisabled = false;
						break;
				}
			}
		}
	},
	findNode: function(event) {
		var target = event.target;
		while (target && !target.syraNodeId) {
			target = target.parentNode;
		}
		if (target && target.syraNodeId) {
			for (var ii = 0, jj = this._nodes.length; ii < jj; ii++) {
				if (this._nodes[ii].syraNodeId == target.syraNodeId) {
					return this._nodes[ii];
				}
			}
		}
		return null;
	},
	selectNode: function(item, select) {
		if (!this._isSelecteDisabled && item && !item.disposed) {
			var node = this._nodes[item.id];
			if (node) {
				if (select) {
					if (this.selectedNode && this.selectedNode != node && this.selectedNode.item) {
						document.site.toggleClass(this.selectedNode.title, "s-aw-tree-designed-" + this.selectedNode.item.$authoringLevel, false);
					}
					this.selectedNode = node;
					document.site.toggleClass(node.title, "s-aw-tree-designed-" + node.item.$authoringLevel, true);
					this.awPalette.scrollToItem(this.selectedNode.node, $(this.slot), true);
				} else {
					if (this.selectedNode == node) {
						this.selectedNode = null;
						document.site.toggleClass(node.title, "s-aw-tree-designed-" + node.item.$authoringLevel, false);
					}
				}
			}
		}
	},
	_addNode: function($item, parentNode) {
		var $prototype = this.awPalette.awArticle.page.$prototype;
		var $items;
		if ($item.$category != "space") {
			if ($item.$items && !$item.$bind && !$item.$category) {
				$items = $item.$items;
			} else {
				if (!parentNode.children) {
					parentNode.children = document.createElement("ul");
					parentNode.children.className = "s-aw-tree-level";
					parentNode.node.appendChild(parentNode.children);
					if (!parentNode.opener) {
						if (parentNode.title) {
							parentNode.opener = document.createElement("a");
							parentNode.opener.className = "s-aw-tree-item-opener";
							parentNode.title.parentNode.insertBefore(parentNode.opener, parentNode.title);
							parentNode.opener.className = "s-aw-tree-item-opener-children";
						}
					}
					if (parentNode.isRoot) {
						parentNode.children.className += " s-aw-tree-root";
					}
				}
				var itemNode = {};
				itemNode.$item = $item;
				itemNode.node = document.createElement("li");
				itemNode.node.syraNodeId = "node-" + this._nodes.length;
				this._nodes.push(itemNode);
				itemNode.node.className = "s-aw-tree-item s-aw-add-item";
				parentNode.children.appendChild(itemNode.node);

				itemNode.data = document.createElement("div");
				itemNode.data.className = "s-aw-tree-item-data";
				itemNode.node.appendChild(itemNode.data);

				itemNode.title = document.createElement("div");
				if (parentNode.isRoot) {
					itemNode.$authoringLevel = "article";
				} else {
					itemNode.$authoringLevel = $item.$bind ? "field" : parentNode.$authoringLevel == "section" ? "block" : "section";
				}
				itemNode.title.className = "s-aw-tree-item-title s-aw-tree-" + itemNode.$authoringLevel;

				var title;
				if (parentNode.isRoot) {
					title = "Available";
				} else {
					if ($item.$title && $item.$title.indexOf("{") >= 0) {
						if ($item.$title[1] == "@") {
							title = $item.$title.slice(1, $item.$title.length - 1);
							if ($prototype.$localization) {
								title = $prototype.$localization[title];
							}
						}
					}
				}
				title = title || $item.$title || "";
				if (title == "-") {
					title = document.site.authorPage.localize["aw_" + itemNode.$authoringLevel];
				}
				var bindTitle = "",
					desc = title;
				if ($item.$bind) {
					bindTitle = this.awPalette.getBindTitle($item.$bind, $prototype.$properties[$item.$bind]);
					if (title == "") {
						desc = title = bindTitle;
					} else {
						desc = title + " (" + bindTitle + ")";
					}
				}
				itemNode.title.textContent = title;
				itemNode.node.title = desc;
				itemNode.data.appendChild(itemNode.title);
			}
			if ($item.$layout && !$item.$bind) {
				$items = $item.$layout.$items;
			}
			if ($items) {
				for (var ii = 0, jj = $items.length; ii < jj; ii++) {
					this._addNode($items[ii], itemNode || parentNode);
				}
			}
		}
	}
});
