// syracuse-ui/lib/authoring/itemPalette.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var RawPage = require("syracuse-ui/lib/page/rawPage").RawPage;
var paletteSettings = require("./setting/paletteSettings");
var palettePrototype = require("./setting/palettePrototype");
var fieldLibrary = require("./field/fieldLibrary");
var historyHelper = require("./historyHelper");
var ToolsBar = require("./aside/toolsBar").ToolsBar;

function ItemPalette(){
}

exports.ItemPalette = helpers.defineClass(ItemPalette, RawPage, {
    notifySteps: function(resetUpdates){
        historyHelper.notifySteps(this.authorPage, this.awContext, resetUpdates);
    },
    onChangeSep: function(menuItem){
        historyHelper.onChangeSep(this.authorPage, this.awContext, menuItem);
    },
    
    loadBox: function(){
        this.awContext = {};
        this._unregisterResizer = true;
        this.authorPage = document.site.authorPage;
        this.localize = this.authorPage.localize;
        this.diagnosePage = this.authorPage.targetPage;
        this.externalAdapter = document.site.externalAdapter;
        this.$prototype = palettePrototype.$prototype;
        this.$autoFetch = false;
        this.$item = paletteSettings.getSettings();
        this.$item.$menus.$isMenusBag = false;
        var keys = Object.keys(this.localize);
        this.$prototype.$localization = {};
        for (var ii = 0, jj = keys.length; ii < jj; ii++) {
            this.$prototype.$localization[keys[ii].replace("aw_", "@")] = this.localize[keys[ii]];
        }
        
        this._initializePage();
        
        this.layoutSlot = document.createElement("div");
        this.authorPage.paletteSlot.appendChild(this.layoutSlot).className = "s-aw-palette-body";
        this.authorPage._paletteBar.setAttribute("data-s-article", this.id);
        RawPage.prototype.loadBox.call(this, {
            $viewType: "layout",
            $links: {
                $help: document.site.$prototype.$links.$help
            }
        });
        this.toolsBar = new ToolsBar();
        this.toolsBar.loadToolbar(this);
        this.notifySteps()// initialize  
        this.authorPage.selectItem(this.authorPage.targetPage, true);
        this.appendArticleMenus(this.authorPage._paletteBar);
    },
    appendLinkBox: function($$slot){
    },
    
    
    onMenuClick: function(menuItem){
        var awItem = this.authorPage.awItem;
        switch (menuItem.$item.$bind) {
            case "$excludeMe":
                this.authorPage.endTargetPageUpdate(this.authorPage.excludeItem(awItem), true);
                break;
            case "$addRowBefore":
            case "$addRowAfter":
                var layout = this.authorPage.awLayout ? this.authorPage.awLayout : (awItem.layoutContent || awItem.layoutParent);
                var newChild = this.authorPage.addSiblingToLayout(layout, menuItem.$item.$bind == "$addRowBefore" ? "top" : "bottom");
                this.authorPage.endTargetPageUpdate(newChild, true);
                break;
            case "$previousLayout":
            case "$nextLayout":
            case "$undoAllLayout":
                this.onChangeSep(menuItem);
                return false;
            case "$help":
                return true;
            default:
                if (menuItem.$item.$bind.indexOf("s-layout-type-") >= 0) {
                    var layout = this.authorPage.awLayout ? this.authorPage.awLayout : awItem.layoutContent;
                    if (awItem.$authoringLevel == "field") {
                        layout = awItem.layoutParent.wrapIntack([awItem], awItem.layoutParent.getItemIndex(awItem, true));
                        layout.removeSpaceBox(true);
                    }
                    var $widths = menuItem.$widths;
                    layout = this.authorPage.newConvert(layout, {
                        $layoutType: $widths == "100" ? "stack" : "row",
                        $widths: $widths
                    });
                    this.authorPage.endTargetPageUpdate(layout.$layout.$layoutType == "stack" ? awItem : layout, true);
                    break;
                }
                awItem.applyDesignMetaData(metaData, true);
                break;
        }
    },
    onWindowResize: function(){
        if (this.layoutSlot) {
            var height = $(this.authorPage.targetPage.scrollview).height();
            if (height) {
                this.layoutSlot.style.height = height + "px";
                this.toolsBar.onWindowResize();
                this.resizeDialogPage();
            }
        }
    },
    resizeDialogPage: function(){
        if (this.authorPage.targetPage.dialogWrapper) {
            var width = this.authorPage.targetPage.dialogWrapper.$$dialogSlot.width();
            if (width) {
                this.authorPage.targetPage.scrollview.style.width = (width - (this.layoutSlot.offsetWidth + this.toolsBar._slot.offsetWidth)) + "px";
            }
        }
    },
    showSettingPanel: function(){
        var selectedPanel, hideStructure;
        var awItem = this.authorPage.awItem;
        var awLayout = this.authorPage.awLayout;
        var paletteId = awItem.isMenuItem ? "menu" : awItem.$authoringLevel;
        for (var ii = 0, jj = this.layoutContent.items.length; ii < jj; ii++) {
            var panel = this.layoutContent.items[ii];
            panel.setState({
                $isHidden: true
            });
            if (panel.$item.$paletteId == paletteId) {
                selectedPanel = panel;
            }
        }
        if (selectedPanel) {
            this.applyChange({
                $isHidden: false,
                $title: "",
                $isMaximizable: false,
                $isBoxCollapsable: false,
                $isTitleHidden: false,
                $fieldsIsTitleHidden: false,
                $labelAlignment: this.authorPage.targetPage.$isEditMode ? "t" : "ll"
            });
            this.selectedPanel = selectedPanel;
            var title = awItem.getTitle(false);
            if (title == "" && awItem.$authoringLevel == "field") {
                title = awItem.$item.$bind;
            }
            this.selectedPanel.setDescription(title);
            
            var $labelAlignment;
            if (awItem.$item.$isTopLabelAlignment || awItem.$item.$fieldsIsTopLabelAlignment) {
                $labelAlignment = "t";
            }
            else {
                if (awItem.$item.$isRightTextLabelAlignment || awItem.$item.$fieldsIsRightTextLabelAlignment) {
                    $labelAlignment = "lr";
                }
                else {
                    $labelAlignment = "ll";
                }
            }
            this.applyChange({
                $title: title,
                $isMaximizable: awItem.$item.$isMaximizable,
                $isBoxCollapsable: awItem.$item.$isBoxCollapsable,
                $isTitleHidden: awItem.$item.$isTitleHidden,
                $fieldsIsTitleHidden: awItem.$item.$fieldsIsTitleHidden,
                $labelAlignment: $labelAlignment,
                $boxTabChoice: (awItem.isTabLayout) ? "tabs" : "stack",
                $layoutTabChoice: (awItem.$item.$layout && awItem.$item.$layout.$layoutType == "tabs") ? "tabs" : "stack"
            });
            this.selectedPanel.setState({
                $isHidden: false
            })
            
            if (this._fieldWidget) {
                this._fieldWidget.designedField = null;
                this.removeItem(this._fieldWidget, true);
                this._fieldWidget = null;
            }
            if (awItem.$authoringLevel == "field") {
                if (awItem.$field) {
                    if (awItem.$field.$type == "application/x-boolean") {
                        var items = this.page.idMap["s-aw-palette-field"].layoutContent.items;
                        for (var ii = 0; ii < jj; ii++) {
                            if (items[ii].$item.$isLabelSection) {
                                items[ii].setState({
                                    $isHidden: true
                                });
                                break;
                            }
                        }
                    }
                    if (this._fieldWidget = fieldLibrary.getAuthoringFieldWidget(awItem)) {
                        this._fieldWidget.layoutSlot = this.idMap["s-aw-field-widget"].$$body[0];
                        this.page.initializeNewItem(this._fieldWidget, {}, this);
                        this._fieldWidget.designedField = awItem;
                        this._fieldWidget.$skin = "s-aw-h2";
                        this._fieldWidget.loadBox(awItem.$item);
                    }
                }
                layout = awItem.layoutParent;
                if (layout && layout.$layout.$layoutType == "row") {
                    hideStructure = true;
                }
            }
            var layout = awLayout || awItem.layoutContent;
            var count = 1;
            if (layout && layout.$item) {
                count = (layout.$item.$widths || "100").split(",").length;
            }
            var metadata = {
                $links: {}
            };
            for (var ii = 0; ii <= 5; ii++) {
                metadata.$links["s-layout-type-" + ii] = {
                    $style: count == ii ? "selected" : null
                };
            }
            metadata.$properties = {
                $layoutTabChoice: {
                    $isDisabled: (count != 1),
                    $title: awItem.$authoringLevel == "article" ? this.localize.aw_showSectionsAs : this.localize.aw_showBlocksAs
                }
            };
            metadata.$properties.$isTitleHidden = {
                $isHidden: awItem.isTabLayout == true
            };
            this.applyChange(metadata);
            var sections = this.selectedPanel.layoutContent.items[0].layoutContent.items;
            var isHidden, section;
            if (sections[0].$bind == "$excludeMe") {
                sections[0].layoutSlot.style.textAlign = "center";
            }
            for (var ii = 0, jj = sections.length; ii < jj; ii++) {
                section = sections[ii];
                isHidden = awLayout == null ? false : true;
                if (section.id == "s-aw-field-widget" && !this._fieldWidget) {
                    isHidden = true;
                }
                if (section.$item.$isRow) {
                    isHidden = awLayout == null ? true : false;
                }
                else {
                    if (hideStructure && section.$item.$isStructure) {
                        isHidden = true;
                    }
                }
                section.setState({
                    $isHidden: isHidden
                });
            }
        }
    },
    notifyDataChange: function(field, value){
        var awItem = this.authorPage.awItem;
        var fieldArticle = field.getArticle();
        if (fieldArticle.designedField) {
            var metaData = {};
            metaData[field.$item.$bind] = value;
            fieldArticle.designedField.applyDesignMetaData(metaData, true);
            fieldArticle.applyChange(metaData);
            if (fieldArticle.onNotifyAuthoringChange) {
                fieldArticle.onNotifyAuthoringChange(field, value, metaData);
            }
            this.authorPage.notifyUpdate();
        }
        else {
            switch (field.$item.$bind) {
                case "$boxTabChoice":
                    if (awItem.layoutParent.$layoutType != value) {
                        this.authorPage.switchTabPanel(awItem, value);
                        this.authorPage.endTargetPageUpdate(awItem, true);
                    }
                    break;
                case "$layoutTabChoice":
                    var children = this.authorPage.getItemsFromLayout(awItem.isLayout ? awItem : awItem.layoutContent);
                    if (children.length > 0 && children[0].$authoringLevel == "field") {
                        break;
                    }
                    this.authorPage.newConvert(awItem.layoutContent, {
                        $layoutType: (value == "tabs") ? "tabs" : "stack"
                    });
                    var children = this.authorPage.getItemsFromLayout(awItem.layoutContent);
                    for (var ii = 0, jj = children.length; ii < jj; ii++) {
                        if (!children[ii].isSpaceBox) {
                            awItem.layoutContent.loadChildItems([children[ii]]);
                        }
                    }
                    var items = [];
                    for (var ii = 0, jj = awItem.layoutContent.items.length; ii < jj; ii++) {
                        if (!awItem.layoutContent.items[ii].isSection) {
                            items.push(awItem.layoutContent.items[ii]);
                        }
                    }
                    for (var ii = 0, jj = items.length; ii < jj; ii++) {
                        awItem.layoutContent.removeItem(items[ii], true);
                    }
                    this.authorPage.endTargetPageUpdate(awItem, true);
                    break;
                case "$viewType":
                    this.authorPage.toggleViewType(value);
                    break;
                case "$labelAlignment":
                    var metaData = {};
                    if (awItem.$authoringLevel == "field") {
                        metaData.$isTopLabelAlignment = (value == "t");
                        if (!metaData.$isTopLabelAlignment) {
                            metaData.$isRightTextLabelAlignment = (value == "lr");
                        }
                        awItem.applyDesignMetaData(metaData, true);
                    }
                    else {
                        if (metaData.$fieldsIsTopLabelAlignment = (value == "t")) {
                            this.authorPage._notifyChildFields(awItem, true, "$isTopLabelAlignment");
                        }
                        else {
                            this.authorPage._notifyChildFields(awItem, false, "$isTopLabelAlignment");
                            this.authorPage._notifyChildFields(awItem, metaData.$fieldsIsRightTextLabelAlignment = (value == "lr"), "$isRightTextLabelAlignment");
                        }
                        this.authorPage._notifyChildBox(awItem, metaData);
                    }
                    this.authorPage.notifyUpdate();
                    break;
                default:
                    var metaData = {};
                    metaData[field.$item.$bind] = value;
                    if (awItem.$authoringLevel == "field") {
                        awItem.applyDesignMetaData(metaData, true);
                    }
                    else {
                        if (field.$item.$bind.indexOf("$fields") == 0) {
                            var fieldBind = field.$item.$bind.slice("$fields".length);
                            fieldBind = fieldBind.slice(0, 1).toLowerCase() + fieldBind.slice(1);
                            this.authorPage._notifyChildFields(awItem, value, "$" + fieldBind);
                        }
                        awItem.applyDesignMetaData(metaData, true);
                    }
                    this.authorPage.notifyUpdate();
                    this.applyChange(metaData);
                    break;
            }
        }
    },
    applyChange: function(newData){
        if (newData) {
            newData.$properties = newData.$properties || {};
            if (newData.$isHidden !== undefined) {
                var keys = ["$title", "$isMaximizable", "$isBoxCollapsable", "$isTitleHidden", "$fieldsIsTitleHidden", "$labelAlignment"];
                for (var ii = 0, jj = keys.length; ii < jj; ii++) {
                    newData.$properties[keys[ii]] = {
                        $isDisabled: newData.$isHidden
                    };
                }
                this.toolsBar.applyChange(newData);
            }
            if (newData.$fieldsIsTitleHidden !== undefined) {
                (newData.$properties.$labelAlignment = newData.$properties.$labelAlignment || {}).$isDisabled = newData.$fieldsIsTitleHidden;
            }
            if (newData.$isTitleHidden !== undefined) {
                newData.$properties.$title = newData.$properties.$title || {};
                newData.$properties.$title.$isDisabled = newData.$isTitleHidden;
                if (this.authorPage.awItem.$authoringLevel == "field") {
                    (newData.$properties.$labelAlignment = newData.$properties.$labelAlignment || {}).$isDisabled = newData.$isTitleHidden;
                }
                else {
                    newData.$properties.$title = newData.$properties.$isBoxCollapsable = newData.$properties.$isMaximizable = {
                        $isDisabled: newData.$isTitleHidden
                    };
                }
            }
        }
        RawPage.prototype.applyChange.call(this, newData);
    },
    dispose: function(){
        if (this._fieldWidget) {
            this._fieldWidget = this._fieldWidget.designedField = null;
        }
        if (this.toolsBar) {
            document.controller.disposeObject(this.toolsBar);
        }
        this.toolsBar = this.diagnosePage = this.$item = this.authorPage = this.localize = this.externalAdapter = this.$prototype = this.layoutSlot = null;
        RawPage.prototype.dispose.call(this);
    }
});
