// syracuse-ui/lib/authoring/itemPalette.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var RawPage = require("syracuse-ui/lib/page/rawPage").RawPage;
var paletteSettings = require("./setting/paletteSettings");
var palettePrototype = require("./setting/palettePrototype");
var fieldLibrary = require("./field/fieldLibrary");
var patternFactory = require("syracuse-ui/lib/authoring/patterns/patternFactory");
var ToolsBar = require("./aside/toolsBar").ToolsBar;
var TreeContent = require("./aside/treeContent").TreeContent;
function ItemPalette(){
}

exports.ItemPalette = helpers.defineClass(ItemPalette, RawPage, {
    onExcludeField: function($bind, isExcluded){
        var $actions = {};
        $actions["$fieldLink" + $bind] = {
            $isHidden: isExcluded
        };
        this.applyChange({
            $actions: $actions
        });
    },
    _refreshStepLinks: function(){
        var metadata;
        if (this.updatedSteps.length > 1) {
            metadata = {
                $links: {
                    $previousLayout: {
                        $title: this.localize.aw_previousLayoutOn.replace("{0}", this._currentUpdtateStep).replace("{1}", this.updatedSteps.length),
                        $isDisabled: this._currentUpdtateStep == 0
                    },
                    $undoAllLayout: {
                        $isDisabled: false
                    }
                }
            };
            if (this._currentUpdtateStep == this.updatedSteps.length - 1) {
                metadata.$links.$nextLayout = {
                    $title: this.localize.aw_nextLayout,
                    $isDisabled: true
                };
            }
            else {
                metadata.$links.$nextLayout = {
                    $title: this.localize.aw_nextLayoutOn.replace("{0}", this._currentUpdtateStep + 1).replace("{1}", this.updatedSteps.length),
                    $isDisabled: false
                };
            }
        }
        else {
            metadata = {
                $links: {
                    $previousLayout: {
                        $title: this.localize.aw_previousLayout,
                        $isDisabled: true
                    },
                    $nextLayout: {
                        $title: this.localize.aw_nextLayout,
                        $isDisabled: true
                    },
                    $undoAllLayout: {
                        $isDisabled: true
                    }
                }
            };
        }
        this.applyChange(metadata);
    },
    _onChangeSep: function(menuItem){
        var self = this;
        var $item;
        switch (menuItem.$item.$bind) {
            case "$previousLayout":
                self._currentUpdtateStep--;
                self.onEndChangeStep(menuItem, $item);
                break;
            case "$nextLayout":
                self._currentUpdtateStep++;
                self.onEndChangeStep(menuItem, $item);
                break;
            case "$undoAllLayout":
                document.site.showMessage({
                    $title: self.localize.aw_updateMessageTitle,
                    $message: self.localize.aw_confirmUndoAll,
                    $type: "question",
                    $buttons: "yesno",
                    callback: function(response){
                        if (response.$id == "yes") {
                            self._currentUpdtateStep = 0;
                            self.notifySteps(true);
                            self.onEndChangeStep(menuItem, $item);
                        }
                    }
                });
                break;
            case "$defaultLayout":
                $item = self.authorPage.$defaultItem;
                self.authorPage.notifyUpdate(true);
                self.onEndChangeStep(menuItem, $item);
                break;
            default:
                if (menuItem.$item.$bind.indexOf("pattern") == 0) {
                    if ((menuItem.$item.$bind != "pattern-headerTabs") && (self.authorPage.awItem.$authoringLevel == "section" || self.authorPage.awItem.$authoringLevel == "block")) {
                        var pattern = patternFactory.patterns[menuItem.$item.$bind];
                        if (pattern.isBoxOnly) {
                            var children = self.authorPage.getItemsFromLayout(self.authorPage.awItem.layoutContent);
                            if (children.length && children[0].$authoringLevel == "field") {
                                return;
                            }
                        }
                        var $result = patternFactory.build(menuItem.$item.$bind, self.authorPage.targetPage.$prototype, self.authorPage.awItem.$item);
                        if (!$result && !$result.$layout) {
                            return;
                        }
                        self.authorPage.awItem.$item.$layout = $result.$layout;
                        $item = self.authorPage.targetPage.$item;
                    }
                    else {
                        $item = patternFactory.build(menuItem.$item.$bind, self.authorPage.targetPage.$prototype, self.authorPage.$sourceItem);
                    }
                    self.authorPage.notifyUpdate(true);
                }
                self.onEndChangeStep(menuItem, $item);
                break;
        }
    },
    onEndChangeStep: function(menuItem, $item){
        var $bind = this.authorPage.awItem.$item.$bind;
        this.authorPage.targetPage.reloadLayout(helpers.object.clone($item || this.updatedSteps[this._currentUpdtateStep], true));
        this._refreshStepLinks();
        var selectItem;
        if ($bind) {
            var bounds = this.authorPage.targetPage.boundFields[$bind];
            selectItem = bounds ? bounds[0] : null;
        }
        this.treeContent.toggle(true);
        this.authorPage.selectItem(selectItem || this.authorPage.targetPage, true);
    },
    notifySteps: function(resetUpdates){
        if (!resetUpdates && this.authorPage.isUpdated) {
            if (this._currentUpdtateStep != this.updatedSteps.length - 1) {
                this.updatedSteps.splice(this._currentUpdtateStep + 1);
                //this.notifySteps(true);
            }
            this.updatedSteps.push(helpers.object.clone(this.authorPage.targetPage.$item, true));
            this._currentUpdtateStep = this.updatedSteps.length - 1;
        }
        else {
            (this.updatedSteps = []).push(this.authorPage.$sourceItem);
            this._currentUpdtateStep = 0;
        }
        this._refreshStepLinks();
    },
    loadBox: function(){
        this._unregisterResizer = true;
        this.authorPage = document.site.authorPage;
        this.localize = this.authorPage.localize;
        this.diagnosePage = this.authorPage.targetPage;
        this.externalAdapter = document.site.externalAdapter;
        this.$prototype = palettePrototype.$prototype;
        this.$autoFetch = false;
        this.$item = paletteSettings.getSettings();
        this.$item.$menus.$isMenusBag = false;
        var keys = Object.keys(this.localize);
        this.$prototype.$localization = {};
        for (var ii = 0, jj = keys.length; ii < jj; ii++) {
            this.$prototype.$localization[keys[ii].replace("aw_", "@")] = this.localize[keys[ii]];
        }
        
        this._initializePage();
        
        this.layoutSlot = document.createElement("div");
        this.authorPage.paletteSlot.appendChild(this.layoutSlot).className = "s-aw-palette-body";
        this.authorPage._paletteBar.setAttribute("data-s-article", this.id);
        RawPage.prototype.loadBox.call(this, {
            $viewType: "layout",
            $links: {
                $help: document.site.$prototype.$links.$help
            }
        });
        this.treeContent = new TreeContent();
        this.treeContent.load();
        this.toolsBar = new ToolsBar();
        this.toolsBar.load(this);
        this.notifySteps()// initialize  
        this.authorPage.selectItem(this.authorPage.targetPage, true);
        this.appendArticleMenus(this.authorPage._paletteBar);
    },
    appendLinkBox: function($$slot){
    },
    excludeItem: function(removedItem){
        if (removedItem.isMenuItem && this.authorPage.targetPage.$pageCategory == "dashboard") {
            this.authorPage.targetPage.garbageLinks[removedItem.$bind] = 1;
        }
        var layoutParent = removedItem.layoutParent;
        var excludedFields = (removedItem.$authoringLevel == "field") ? [removedItem] : this.authorPage._extractFields(removedItem.layoutContent);
        if (excludedFields) {
            for (var ii = 0, jj = excludedFields.length; ii < jj; ii++) {
                this.onExcludeField(excludedFields[ii].$item.$bind, false);
            }
        }
        var index = layoutParent.getItemIndex(removedItem, true);
        layoutParent.removeItem(removedItem, true);
        var newItem = layoutParent.items[index];
        if (!newItem) {
            newItem = layoutParent.items[--index];
        }
        newItem = newItem || layoutParent.boxParent;
        switch (layoutParent.$layout.$layoutType) {
            case "row":
                layoutParent.addSpaceBox(index);
                break;
            case "tabs":
                if (layoutParent.items.length == 0) {
                    this.authorPage.newConvert(layoutParent, {
                        $layoutType: "stack",
                        $width: "100"
                    });
                }
                break;
        }
        if (this.treeContent) {
            this.treeContent.toggle(true);
        }
        return newItem;
    },
    onMenuClick: function(menuItem){
        var awItem = this.authorPage.awItem;
        switch (menuItem.$item.$bind) {
            case "$excludeMe":
                this.authorPage.endTargetPageUpdate(this.excludeItem(awItem), true);
                break;
            case "$addRowBefore":
            case "$addRowAfter":
                var layout = this.authorPage.awLayout ? this.authorPage.awLayout : (awItem.layoutContent || awItem.layoutParent);
                var newChild = this.authorPage.addSiblingToLayout(layout, menuItem.$item.$bind == "$addRowBefore" ? "top" : "bottom");
                this.authorPage.endTargetPageUpdate(newChild, true);
                break;
            case "$addSection":
            case "$addBlock":
                var newChild = this.authorPage.addNewItem(awItem, menuItem.$item.$bind.replace("$add", "").toLowerCase());
                this.authorPage.endTargetPageUpdate(newChild, true);
                return false;
            case "$addMenus":
                var newChild = this.authorPage.addNewItem(awItem, "block", null, undefined, undefined, "menus");
                this.authorPage.endTargetPageUpdate(newChild, true);
                return false;
            case "$previousLayout":
            case "$nextLayout":
            case "$undoAllLayout":
            case "$defaultLayout":
            case "pattern-headerTabs":
                this._onChangeSep(menuItem);
                return false;
            case "$help":
                return true;
            default:
                if (menuItem.$item.$fieldBind) {
                    this.authorPage.onAddNewField(awItem, menuItem.$item);                
                }
                else {
                    if (menuItem.$item.$bind.indexOf("s-layout-type-") >= 0) {
                        var layout = this.authorPage.awLayout ? this.authorPage.awLayout : awItem.layoutContent;
                        if (awItem.$authoringLevel == "field") {
                            layout = awItem.layoutParent.wrapIntack([awItem], awItem.layoutParent.getItemIndex(awItem, true));
                            layout.removeSpaceBox(true);
                        }
                        var $widths = menuItem.$widths;
                        layout = this.authorPage.newConvert(layout, {
                            $layoutType: $widths == "100" ? "stack" : "row",
                            $widths: $widths
                        });
                        this.authorPage.endTargetPageUpdate(layout.$layout.$layoutType == "stack" ? awItem : layout, true);
                        break;
                    }
                    if (menuItem.$item.$bind.indexOf("pattern") == 0) {
                        this._onChangeSep(menuItem);
                        break;
                    }
                    awItem.applyDesignMetaData(metaData, true);
                }
                break;
        }
    },
    onWindowResize: function(){
        if (this.layoutSlot) {
            var height = $(this.authorPage.targetPage.scrollview).height();
            if (height) {
                this.layoutSlot.style.height = height + "px";
                this.toolsBar.onWindowResize();
                this.resizeDialogPage();
            }
        }
    },
    resizeDialogPage: function(){
        if (this.authorPage.targetPage.dialogWrapper) {
            var width = this.authorPage.targetPage.dialogWrapper.$$dialogSlot.width();
            if (width) {
                this.authorPage.targetPage.scrollview.style.width = (width - (this.layoutSlot.offsetWidth + this.toolsBar._slot.offsetWidth)) + "px";
            }
        }
    },
    showSettingPanel: function(){
        var selectedPanel, hideStructure;
        var awItem = this.authorPage.awItem;
        var awLayout = this.authorPage.awLayout;
        var paletteId = awItem.isMenuItem ? "menu" : awItem.$authoringLevel;
        for (var ii = 0, jj = this.layoutContent.items.length; ii < jj; ii++) {
            var panel = this.layoutContent.items[ii];
            panel.setState({
                $isHidden: true
            });
            if (panel.$item.$paletteId == paletteId) {
                selectedPanel = panel;
            }
        }
        if (selectedPanel) {
            this.applyChange({
                $isHidden: false,
                $title: "",
                $isMaximizable: false,
                $isBoxCollapsable: false,
                $isTitleHidden: false,
                $fieldsIsTitleHidden: false,
                $labelAlignment: this.authorPage.targetPage.$isEditMode ? "t" : "ll"
            });
            this.selectedPanel = selectedPanel;
            var title = awItem.getTitle(false);
            if (title == "" && awItem.$authoringLevel == "field") {
                title = awItem.$item.$bind;
            }
            this.selectedPanel.setDescription(title);
            
            var $labelAlignment;
            if (awItem.$item.$isTopLabelAlignment || awItem.$item.$fieldsIsTopLabelAlignment) {
                $labelAlignment = "t";
            }
            else {
                if (awItem.$item.$isRightTextLabelAlignment || awItem.$item.$fieldsIsRightTextLabelAlignment) {
                    $labelAlignment = "lr";
                }
                else {
                    $labelAlignment = "ll";
                }
            }
            this.applyChange({
                $title: title,
                $isMaximizable: awItem.$item.$isMaximizable,
                $isBoxCollapsable: awItem.$item.$isBoxCollapsable,
                $isTitleHidden: awItem.$item.$isTitleHidden,
                $fieldsIsTitleHidden: awItem.$item.$fieldsIsTitleHidden,
                $labelAlignment: $labelAlignment,
                $boxTabChoice: (awItem.isTabLayout) ? "tabs" : "stack",
                $layoutTabChoice: (awItem.$item.$layout && awItem.$item.$layout.$layoutType == "tabs") ? "tabs" : "stack"
            });
            this.selectedPanel.setState({
                $isHidden: false
            })
            
            if (this._fieldWidget) {
                this._fieldWidget.designedField = null;
                this.removeItem(this._fieldWidget, true);
                this._fieldWidget = null;
            }
            if (awItem.$authoringLevel == "field") {
                if (awItem.$field) {
                    if (awItem.$field.$type == "application/x-boolean") {
                        var items = this.page.idMap["s-aw-palette-field"].layoutContent.items;
                        for (var ii = 0; ii < jj; ii++) {
                            if (items[ii].$item.$isLabelSection) {
                                items[ii].setState({
                                    $isHidden: true
                                });
                                break;
                            }
                        }
                    }
                    if (this._fieldWidget = fieldLibrary.getAuthoringFieldWidget(awItem)) {
                        this._fieldWidget.layoutSlot = this.idMap["s-aw-field-widget"].$$body[0];
                        this.page.initializeNewItem(this._fieldWidget, {}, this);
                        this._fieldWidget.designedField = awItem;
                        this._fieldWidget.$skin = "s-aw-h2";
                        this._fieldWidget.loadBox(awItem.$item);
                    }
                }
                layout = awItem.layoutParent;
                if (layout && layout.$layout.$layoutType == "row") {
                    hideStructure = true;
                }
            }
            var layout = awLayout || awItem.layoutContent;
            var count = 1;
            if (layout && layout.$item) {
                count = (layout.$item.$widths || "100").split(",").length;
            }
            var metadata = {
                $links: {}
            };
            for (var ii = 0; ii <= 5; ii++) {
                metadata.$links["s-layout-type-" + ii] = {
                    $style: count == ii ? "selected" : null
                };
            }
            metadata.$properties = {
                $layoutTabChoice: {
                    $isDisabled: (count != 1),
                    $title: awItem.$authoringLevel == "article" ? this.localize.aw_showSectionsAs : this.localize.aw_showBlocksAs
                }
            };
            metadata.$properties.$isTitleHidden = {
                $isHidden: awItem.isTabLayout == true
            };
            this.applyChange(metadata);
            var sections = this.selectedPanel.layoutContent.items[0].layoutContent.items;
            var isHidden, section;
            if (sections[0].$bind == "$excludeMe") {
                sections[0].layoutSlot.style.textAlign = "center";
            }
            for (var ii = 0, jj = sections.length; ii < jj; ii++) {
                section = sections[ii];
                isHidden = awLayout == null ? false : true;
                if (section.id == "s-aw-field-widget" && !this._fieldWidget) {
                    isHidden = true;
                }
                if (section.$item.$isRow) {
                    isHidden = awLayout == null ? true : false;
                }
                else {
                    if (hideStructure && section.$item.$isStructure) {
                        isHidden = true;
                    }
                }
                section.setState({
                    $isHidden: isHidden
                });
            }
        }
    },
    notifyDataChange: function(field, value){
        var awItem = this.authorPage.awItem;
        var fieldArticle = field.getArticle();
        if (fieldArticle.designedField) {
            var metaData = {};
            metaData[field.$item.$bind] = value;
            fieldArticle.designedField.applyDesignMetaData(metaData, true);
            fieldArticle.applyChange(metaData);
            if (fieldArticle.onNotifyAuthoringChange) {
                fieldArticle.onNotifyAuthoringChange(field, value, metaData);
            }
            this.authorPage.notifyUpdate();
        }
        else {
            switch (field.$item.$bind) {
                case "$boxTabChoice":
                    if (awItem.layoutParent.$layoutType != value) {
                        this.authorPage.switchTabPanel(awItem, value);
                        this.authorPage.endTargetPageUpdate(awItem, true);
                    }
                    break;
                case "$layoutTabChoice":
                    var children = this.authorPage.getItemsFromLayout(awItem.isLayout ? awItem : awItem.layoutContent);
                    if (children.length > 0 && children[0].$authoringLevel == "field") {
                        break;
                    }
                    this.authorPage.newConvert(awItem.layoutContent, {
                        $layoutType: (value == "tabs") ? "tabs" : "stack"
                    });
                    var children = this.authorPage.getItemsFromLayout(awItem.layoutContent);
                    for (var ii = 0, jj = children.length; ii < jj; ii++) {
                        if (!children[ii].isSpaceBox) {
                            awItem.layoutContent.loadChildItems([children[ii]]);
                        }
                    }
                    var items = [];
                    for (var ii = 0, jj = awItem.layoutContent.items.length; ii < jj; ii++) {
                        if (!awItem.layoutContent.items[ii].isSection) {
                            items.push(awItem.layoutContent.items[ii]);
                        }
                    }
                    for (var ii = 0, jj = items.length; ii < jj; ii++) {
                        awItem.layoutContent.removeItem(items[ii], true);
                    }
                    this.authorPage.endTargetPageUpdate(awItem, true);
                    break;
                case "$viewType":
                    this.authorPage.toggleViewType(value);
                    break;
                case "$labelAlignment":
                    var metaData = {};
                    if (awItem.$authoringLevel == "field") {
                        metaData.$isTopLabelAlignment = (value == "t");
                        if (!metaData.$isTopLabelAlignment) {
                            metaData.$isRightTextLabelAlignment = (value == "lr");
                        }
                        awItem.applyDesignMetaData(metaData, true);
                    }
                    else {
                        if (metaData.$fieldsIsTopLabelAlignment = (value == "t")) {
                            this.authorPage._notifyChildFields(awItem, true, "$isTopLabelAlignment");
                        }
                        else {
                            this.authorPage._notifyChildFields(awItem, false, "$isTopLabelAlignment");
                            this.authorPage._notifyChildFields(awItem, metaData.$fieldsIsRightTextLabelAlignment = (value == "lr"), "$isRightTextLabelAlignment");
                        }
                        this.authorPage._notifyChildBox(awItem, metaData);
                    }
                    this.authorPage.notifyUpdate();
                    break;
                default:
                    var metaData = {};
                    metaData[field.$item.$bind] = value;
                    if (awItem.$authoringLevel == "field") {
                        awItem.applyDesignMetaData(metaData, true);
                    }
                    else {
                        if (field.$item.$bind.indexOf("$fields") == 0) {
                            var fieldBind = field.$item.$bind.slice("$fields".length);
                            fieldBind = fieldBind.slice(0, 1).toLowerCase() + fieldBind.slice(1);
                            this.authorPage._notifyChildFields(awItem, value, "$" + fieldBind);
                        }
                        awItem.applyDesignMetaData(metaData, true);
                    }
                    this.authorPage.notifyUpdate();
                    this.applyChange(metaData);
                    break;
            }
        }
    },
    applyChange: function(newData){
        if (newData) {
            newData.$properties = newData.$properties || {};
            if (newData.$isHidden !== undefined) {
                var keys = ["$title", "$isMaximizable", "$isBoxCollapsable", "$isTitleHidden", "$fieldsIsTitleHidden", "$labelAlignment"];
                for (var ii = 0, jj = keys.length; ii < jj; ii++) {
                    newData.$properties[keys[ii]] = {
                        $isDisabled: newData.$isHidden
                    };
                }
                keys = ["$addBlock", "$addSection", "$addField", "$addMenus"];
                for (var ii = 0, jj = keys.length; ii < jj; ii++) {
                    if (this.menuItems[keys[ii]]) {
                        this.menuItems[keys[ii]][0].setMenu({
                            $isDisabled: newData.$isHidden
                        });
                    }
                }
            }
            if (newData.$fieldsIsTitleHidden !== undefined) {
                (newData.$properties.$labelAlignment = newData.$properties.$labelAlignment || {}).$isDisabled = newData.$fieldsIsTitleHidden;
            }
            if (newData.$isTitleHidden !== undefined) {
                newData.$properties.$title = newData.$properties.$title || {};
                newData.$properties.$title.$isDisabled = newData.$isTitleHidden;
                if (this.authorPage.awItem.$authoringLevel == "field") {
                    (newData.$properties.$labelAlignment = newData.$properties.$labelAlignment || {}).$isDisabled = newData.$isTitleHidden;
                }
                else {
                    newData.$properties.$title = newData.$properties.$isBoxCollapsable = newData.$properties.$isMaximizable = {
                        $isDisabled: newData.$isTitleHidden
                    };
                }
            }
        }
        RawPage.prototype.applyChange.call(this, newData);
    },
    dispose: function(){
        if (this._fieldWidget) {
            this._fieldWidget = this._fieldWidget.designedField = null;
        }
        if (this.toolsBar) {
            this.toolsBar.dispose();
        }
        if (this.treeContent) {
            this.treeContent.dispose();
        }
        this.treeContent = this.toolsBar = this.diagnosePage = this.$item = this.authorPage = this.localize = this.externalAdapter = this.$prototype = this.layoutSlot = null;
        RawPage.prototype.dispose.call(this);
    }
});
