// syracuse-ui/lib/authoring/landingPageDesigner.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Designer = require("syracuse-ui/lib/authoring/designer").Designer;

function LandingPageDesigner() {}

exports.LandingPageDesigner = helpers.defineClass(LandingPageDesigner, Designer, {
	drawBox: function() {
		this.layoutSlot = document.createElement("aside");
		this.layoutSlot.id = "s-aw-page";

		this.awArticle.header.appendChild(this.layoutSlot);
		this.layoutSlot.style.display = "";
		this.$$item = $(this.layoutSlot);
		this.sectionModel = {
			body: this.layoutSlot.appendChild(document.createElement("div"))
		};
		this.sectionModel.body.className = "s-aw-lp-model-slot";
		//right part
		this.awItem = this.awArticle;

		this.loadDefaultTools();
		this.workinCopySlot.appendChild(this.historySlot);
		this.layoutSlot.appendChild(this.workinCopySlot);

		this.loadTemplateTools();
		this.templatesTool.addModelLinks(this.sectionModel, this.awItem.layoutContent.tabOpened);
		document.site.resize();
	},
	dispose: function() {
		document.site.removeDomChild(this.layoutSlot);
		this.sectionModel = null;
		Designer.prototype.dispose.call(this);
	}
});
