// syracuse-ui/lib/authoring/pageDesigner.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Designer = require("syracuse-ui/lib/authoring/designer").Designer;
var BarSplitter = require("syracuse-ui/lib/page/bar/barSplitter").BarSplitter;
var InsertTree = require("syracuse-ui/lib/authoring/tree/insertTree").InsertTree;
var ContentTree = require("syracuse-ui/lib/authoring/tree/contentTree").ContentTree;
var _designerViewChoice = require('syracuse-ui/lib/authoring/designerViewChoice');
var _uiSwitchTool = require("syracuse-ui/lib/authoring/tools/uiSwitchTool");
var _localizationTool = require("syracuse-ui/lib/authoring/tools/localizationTool");

var PageStorage = require("syracuse-ui/lib/authoring/pageStorage").PageStorage;
var ItemTool = require("syracuse-ui/lib/authoring/tools/itemTool").ItemTool;

var _settingPanel = require("syracuse-ui/lib/authoring/tools/settingPanel");
var _fieldAddinTypes = [require("syracuse-ui/lib/authoring/fieldAddins/choiceFieldAddin"), require("syracuse-ui/lib/authoring/fieldAddins/documentFieldAddin")];

function _onDesignerApplyChangeToLabel(designedItem, newData) {
	var $titleChoice = newData.$fieldsTitleChoice || newData.$fieldTitleChoice;
	if ($titleChoice !== undefined) {
		if (!newData.$properties.$labelAlignment) {
			newData.$properties.$labelAlignment = {};
		}
		newData.$properties.$labelAlignment.$isDisabled = $titleChoice == 2;
	}
	if (newData.$isTitleHidden !== undefined) {
		newData.$properties.$title = newData.$properties.$title || {};
		newData.$properties.$title.$isDisabled = newData.$isTitleHidden;
		newData.$properties.$title = newData.$properties.$isBoxCollapsable = newData.$properties.$isMaximizable = {
			$isDisabled: newData.$isTitleHidden
		};
	}
}

function _getTitleChoiceProto() {
	return {
		$type: "application/x-choice",
		$isTitleHidden: true,
		$value: {
			$type: "application/x-string",
			$enum: [{
				$value: 0,
				$title: syra_local.aw_titleVisible
			}, {
				$value: 1,
				$title: syra_local.aw_titleBlank
			}, {
				$value: 2,
				$title: syra_local.aw_titleHidden
			}]
		}
	};
}

function _addLabelFields(designer, $stepSkin) {
	designer.$prototype.$properties.$labelAlignment = {
		$type: "application/x-choice",
		$title: syra_local.aw_labelPosition,
		$value: {
			$type: "application/x-string",
			$enum: [{
				$value: "t",
				$title: syra_local.aw_labelPositionTop
			}, {
				$value: "ll",
				$title: syra_local.aw_labelPositionLeftLeft
			}, {
				$value: "lr",
				$title: syra_local.aw_labelPositionLeftRight
			}]
		}
	};
	designer.$prototype.$properties.$fieldsTitleChoice = _getTitleChoiceProto();
	designer.$prototype.$properties.$fieldTitleChoice = _getTitleChoiceProto();
}

function _addViewBtns(designer, slot) {
	var types = ["viewTypePreview", "viewTypeLayout", "viewTypeStructure"];
	syra_site.mobileGateway && types.push("viewTypeJSON");

	designer.onViewClick = function(event, btn) {
		var designer = this.parent;
		if (designer.$viewType) {
			syra_site.dom.toggleClass(designer[designer.$viewType + "_btn"].link, "s-selected", false);
		}
		designer.$viewType = this.viewType;
		if (designer.isJsonMode && designer.$viewType != "viewTypeJSON") {
			designer.isJsonMode = false;
			syra_site.mobileGateway.toggleJsonEditor(false);
		}
		_ensureDesigneBar(designer);
		if (designer.$viewType == "viewTypeJSON") {
			designer.isJsonMode = true;
			syra_site.mobileGateway.toggleJsonEditor(true);
		}
		syra_site.dom.toggleClass(this.link, "s-selected", true);
		designer.toggleUIDesign(designer.designedArticle, true);
		syra_site.ensureArticleVisibility(designer.designedArticle.page);
	};

	types.forEach(function(type) {
		designer[type + "_btn"] = syra_menus.button.add({
			parent: designer,
			slot: slot,
			text: syra_local["aw_" + type],
			css: "s-aw-viewType-link",
			btnclick: designer.onViewClick,
			viewType: type
		});
	});

	syra_site.dom.toggleClass(designer[(designer.$viewType = "viewTypeLayout") + "_btn"].link, "s-selected", true);

};

function _selectTreesTab(designer, tab, select) {
	if (select && designer._selectedTab && designer._selectedTab.tab != tab) {
		_selectTreesTab(designer, designer._selectedTab.tab, false);
	}
	var tabItem = designer.contentTree.tabItem.tab == tab ? designer.contentTree.tabItem : designer.insertTree.tabItem;
	designer._selectedTab = (select) ? tabItem : null;
	tabItem.tab.className = "new-s-aw-tlbr-tab" + (select ? " s-tab-open" : "");
	tabItem.body.style.display = select ? "" : "none";
}

function _addTreesTab(designer, title) {
	var body = document.createElement("div");
	body.style.display = "none";
	body.className = "s-aw-content-tree";
	return {
		tab: syra_menus.button.add({
			parent: designer,
			slot: designer.treebarTabs,
			text: title,
			css: "new-s-aw-tlbr-tab",
			btnclick: function() {
				_selectTreesTab(this.parent, this.link, true);
			}
		}).link,
		body: designer.treebarBody.appendChild(body)
	};
}

function _addTreeBar(designer) {
	designer.treebarBody = document.createElement("div");
	designer.treebarBody.className = "s-aw-treebar-body";

	var bar = designer.treesBar = new BarSplitter();
	bar.options = {
		resizeDirection: "left",
		$skin: "s-aw-treebar"
	};
	bar.resizeBar = function() {
		if (BarSplitter.prototype.resizeBar.call(this)) {
			if (designer.contentTree) {
				var height = syra_site.dom.getInnerSize(this.slot).height;
				var bodyHeight = Math.max(height - designer.treebarTabs.getBoundingClientRect().height, 20) + "px";
				designer.contentTree.tabItem.body.style.height = bodyHeight;
				designer.insertTree.tabItem.body.style.height = bodyHeight;
			}
		}
	};
	bar.isDesignerBar = true;
	bar.layoutSlot = designer.propertiesSlot;
	bar.slot = designer.treebarSlot;
	bar.body = designer.treebarBody;
	bar.load(designer.designedArticle.page);


	designer.fillTree = function() {
		if (this.designedArticle) {
			this.contentTree.fill();
			this.insertTree.fill();
			if (this._selectedTab && this._selectedTab.tab) {
				_selectTreesTab(this, this._selectedTab.tab, true);
			}
		}
	};
	designer.treebarTabs = document.createElement("div");
	designer.treebarTabs.className = "s-aw-tlbr-tabs-nav";
	designer.treebarBody.appendChild(designer.treebarTabs);
	designer._selectedTab = null;
	designer.designedArticle.garbage.loadFreeField();
	(designer.contentTree = new ContentTree()).load(designer, _addTreesTab(designer, syra_local.aw_viewTypeContent));
	(designer.insertTree = new InsertTree()).load(designer, _addTreesTab(designer, syra_local.aw_insert));
	_selectTreesTab(designer, designer.contentTree.tabItem.tab, true);
}

function _ensureDesigneBar(designer, isDesigning) {
	isDesigning = isDesigning !== false;
	if (isDesigning) {
		designer.isDesignBarVisible = !(designer.$viewType == "viewTypePreview");
		designer.propertiesSlot.style.display = designer.isDesignBarVisible ? "" : "none";
	} else {
		designer.isDesignBarVisible = false;
	}

	designer.asideParent.domItem.style.paddingLeft = designer.isDesignBarVisible ? (designer.treebarSlot.clientWidth + "px") : "";
	designer.asideParent.domItem.style.paddingRight = designer.isDesignBarVisible ? (designer.propertiesSlot.clientWidth + "px") : "";

	var page = designer.designedArticle;
	for (var ii = 0, jj = page.bars.length; ii < jj; ii++) {
		var bar = page.bars[ii];
		var visible = designer.isDesignBarVisible == !! bar.isDesignerBar;
		if (visible) {
			if (bar.beforeDesign) {
				bar.beforeDesign.parentNode && bar.beforeDesign.parentNode.insertBefore(bar.slot, bar.slot.nextSibling);
				delete bar.beforeDesign;
				bar.resizeSplitter();
			}
		} else {
			if (!bar.beforeDesign) {
				bar.beforeDesign = {
					parentNode: bar.slot.parentNode,
					nextSibling: bar.slot.nextSibling
				};
				syra_site.dom.removeChild(bar.slot);
			}

		}
	}
}

var _stack = [];

exports.Designer = helpers.defineClass(function() {}, Designer, {
	setFieldTitleItem: function(item) {
		var $titleChoice = item.$item.$isTitleHidden ? 2 : ((item.$item.$isTitleEmpty) ? 1 : 0);
		this.applyChange({
			$fieldTitleChoice: $titleChoice,
		});
	},
	isTopPageDesigner: function() {
		return _stack.indexOf(this) == (_stack.length - 1);
	},
	loadBox: function() {

		(_stack.indexOf(this) < 0) && _stack.push(this);

		this.$prototype = {
			$properties: {
				$isMaximizable: {
					$type: "application/x-boolean",
					$title: syra_local.aw_isMaximizable
				},
				$isBoxCollapsable: {
					$type: "application/x-boolean",
					$title: syra_local.aw_isBoxCollapsable
				},
				$isAdvanced: {
					$type: "application/x-boolean",
					$title: syra_local.aw_isAdvanced
				},

				$rowAlign: {
					$type: "application/x-choice",
					$title: syra_local.aw_rowAlign,
					$value: {
						$type: "application/x-string",
						$enum: [{
							$value: "top",
							$title: syra_local.aw_rowAlignTop
						}, {
							$value: "bottom",
							$title: syra_local.aw_rowAlignBottom
						}]
					}
				}
			}
		};
		this.panels = {};
		Designer.prototype.loadBox.call(this);
	},
	drawBox: function() {
		syra_site.registerScreenDesigner(true, this);
		(this.itemTool = new ItemTool()).load(this);

		this.domItem = this.layoutSlot = document.createElement("aside");
		this.layoutSlot.className = "s-aw-page";
		syra_site.header.parentNode.insertBefore(this.layoutSlot, syra_site.header.nextSibling);
		this.layoutSlot.style.display = "";

		//left part
		this.leftTop = document.createElement("div");
		this.leftTop.className = "s-aw-page-left";
		this.layoutSlot.appendChild(this.leftTop);

		//middle part
		var viewsSlot = this.layoutSlot.appendChild(document.createElement("div"));
		viewsSlot.className = "s-aw-page-views";

		//right part
		this.propertiesSlot = document.createElement("div");
		this.propertiesSlot.className = "s-aw-properties-slot";
		this.propertiesSlot.style.width = "200px";
		this.setArticleId(this.propertiesSlot);

		this.treebarSlot = document.createElement("div");
		this.treebarSlot.className = "s-aw-treebar-slot";

		this.asideParent = this.designedArticle.page;
		if (this.asideParent.isVignettePage) {
			this.asideParent = this.asideParent.getMasterPage();
		}
		this.asideParent.domItem.appendChild(this.propertiesSlot);
		this.asideParent.domItem.appendChild(this.treebarSlot);
		this.setArticleId(this.treebarSlot);
		_uiSwitchTool.load(this);
		if (this.designedArticle == this.designedArticle.page) {
			_addViewBtns(this, viewsSlot);
		}
		this.designerHelpLink = this.loadNewItem(this.leftTop, {
			$bind: "$designerDocumentation",
			$category: "link",
			$noText: true,
			$skin: "s-design-history-btn-icon",
			$icon: {
				$value: "help"
			}
		}, syra_site);

		this.loadHistoryTool();

		this.loadPageStorage();

		_localizationTool.load(this);
		this.loadTemplateTools();
		_addLabelFields(this);

		for (var ii = 0, jj = _fieldAddinTypes.length; ii < jj; ii++) {
			_fieldAddinTypes[ii].load(this);
		}

		_addTreeBar(this);
		this.propertiesSlot.style.display = "";
		this.treebarSlot.style.display = "";

		this.toggleUIDesign(this.designedArticle, true);

		if (this.designedArticle.$item.$transaction && this.leftTop) {
			var $title = "Transaction";
			if (this.designedArticle.$item.$isModel) {
				$title += " Model";
			}
			this.transactionLabel = document.createElement("div");
			this.transactionLabel.style.display = "";
			this.transactionLabel.textContent = $title;
			this.transactionLabel.id = "s-aw-transaction";
			this.transactionLabel.style.display = "none";
			this.leftTop.appendChild(this.transactionLabel);
		}
		_ensureDesigneBar(this);
		this.selectItem(this.designedArticle, true);
	},
	addFieldAddinToSection: function(section) {
		for (var ii = 0, jj = _fieldAddinTypes.length; ii < jj; ii++) {
			_fieldAddinTypes[ii].addToSection(this, section);
		}
	},
	onShowFieldAddins: function() {
		for (var ii = 0, jj = _fieldAddinTypes.length; ii < jj; ii++) {
			_fieldAddinTypes[ii].onShow(this);
		}
	},
	displayViewTitle: function($view) {
		if ($view) {
			this.viewTitleSlot.textContent = $view.$isFactory ? syra_local.aw_pg_choice_factory : $view.$title;
			this.viewTitleSlot.style.display = "";
		} else {
			this.viewTitleSlot.style.display = "none";
		}
	},
	onEndHistoryChangeStep: function(itemToSelect) {
		this.fillTree && this.fillTree();
		this.toggleUIDesign(this.designedArticle, true);
		this.selectItem(itemToSelect, true);
	},
	endArticleUpdate: function(selectedItem, isStructureUpdated) {
		syra_site.ensureArticleVisibility(this.designedArticle.page);
		isStructureUpdated && this.fillTree && this.fillTree();
		selectedItem = selectedItem || this.designedItem;
		this.selectItem(selectedItem, true);
		this.toggleUIDesign(this.designedArticle, true);
		this.history.updateSteps();
		this.saveDesign();
	},
	saveDesign: function() {
		if (this.pageStorage) {
			this.pageStorage.saveDesign();
		}
	},
	onBeforeMainPageChange: function(continueChanging) {
		if (this.pageStorage && this.pageStorage.onBeforeMainPageChange(continueChanging) == false) {
			return false;
		}
		this.closeDesigner();
		return true;
	},
	openDesigner: function(designedArticle) {
		Designer.prototype.openDesigner.call(this, designedArticle);
		_designerViewChoice.disableOpener(this.designedArticle, true);
	},
	closeDesigner: function() {
		_designerViewChoice.disableOpener(this.designedArticle, false);
		syra_site.switchItemDesigner(this.designedArticle, false);
	},
	onTopScreenDesigner: function(isTop) {
		this.propertiesSlot.style.display = this.layoutSlot.style.display = this.treebarSlot.style.display = isTop ? "" : "none";
	},
	loadPageStorage: function() {
		var id;
		if (this.pageStorage) {
			id = this.pageStorage.id;
			this.pageStorage.dispose();

		}
		this.pageStorage = new PageStorage();
		this.pageStorage.id = id;
		this.pageStorage.load(this);
	},


	selectItem: function(item, select) {
		if (item && item.isMenuItem && item.referenceField) {
			item = item.referenceField;
		}
		if (this.awLayout && this.awLayout.disposed) {
			this.awLayout = null;
		}
		if (this.designedItem && this.designedItem.disposed) {
			this.designedItem = null;
		}
		if (item && item.disposed) {
			item = this.designedItem;
		}
		if (select) {
			item = item || this.designedArticle.page;
			this.selectItem(null, false);
			if (item.isLayout) {
				this.designedItem = item.boxParent;
				this.awLayout = item;
			} else {
				this.designedItem = item;
				this.awLayout = null;
			}

			this.designedArticle.page.scrollToItem(this.awLayout || this.designedItem);
			_settingPanel.show(this);
			this.awLayout && this.toggleItemCss(this.awLayout, "s-aw-designed", true);
			this.toggleItemCss(item, "s-aw-designed", true);
		} else {
			item = item || this.designedItem;
			item && this.toggleItemCss(item, "s-aw-designed", false);
			this.awLayout && this.toggleItemCss(this.awLayout, "s-aw-designed", false);
		}
		if (this.contentTree) {
			this.contentTree.selectNode(item, select);
			this.insertTree.selectNode(item, select);
		}
	},
	applyChange: function(newData) {
		if (newData) {
			newData.$properties = newData.$properties || {};
			if (newData.$isHidden !== undefined) {
				newData.$properties.$isMaximizable = {
					$isDisabled: newData.$isHidden
				};
				newData.$properties.$isBoxCollapsable = {
					$isDisabled: newData.$isHidden
				};
			}
			_onDesignerApplyChangeToLabel(this.designedItem, newData);
			for (var ii = 0, jj = _fieldAddinTypes.length; ii < jj; ii++) {
				_fieldAddinTypes[ii].onDesignerApplyChange(this, newData);
			}
			if (newData.$title !== undefined) {
				this._ensureSelectedItemTitle();
			}
		}
		Designer.prototype.applyChange.call(this, newData);
	},
	notifyDataChange: function(field, value) {
		this.saveDesign();
		switch (field.$item.$bind) {
			case "$rowAlign":
				if (this.awLayout) {
					this.awLayout.$layout.$rowAlign = value;
					for (var ii = 0, jj = this.awLayout.items.length; ii < jj; ii++) {
						this.awLayout.items[ii].layoutSlot.style.verticalAlign = value;
					}
					this.endArticleUpdate(this.awLayout, true);
				}
				break;
			default:
				var metaData = {};
				if (!_localizationTool.onNotifyDataChange(this, metaData, field, value)) {
					switch (field.$item.$bind) {
						case "$fieldsTitleChoice":
							var $meta = {
								$fieldsIsTitleEmpty: false,
								$fieldsIsTitleHidden: false
							};
							switch (value) {
								case 1:
									$meta.$fieldsIsTitleEmpty = true;
									break;
								case 2:
									$meta.$fieldsIsTitleHidden = true;
									break;
							}
							this.designedItem.applyDesignMetaData($meta, true);
							metaData[field.$item.$bind] = value;
							break;
						case "$fieldTitleChoice":
							var $meta = {
								$isTitleEmpty: false,
								$isTitleHidden: false
							};
							switch (value) {
								case 1:
									$meta.$isTitleEmpty = true;
									break;
								case 2:
									$meta.$isTitleHidden = true;
									break;
							}
							this.designedItem.applyDesignMetaData($meta, true);
							metaData[field.$item.$bind] = value;
							break;
						case "$labelAlignment":
							metaData.$isTopLabelAlignment = (value == "t");
							if (metaData.$isTopLabelAlignment == false) {
								metaData.$isRightTextLabelAlignment = (value == "lr");
							}
							if (this.designedItem.isField) {
								this.designedItem.applyDesignMetaData(metaData, true);
							} else {
								this.designedItem.applyDesignMetaData({
									$fieldsIsTopLabelAlignment: metaData.$isTopLabelAlignment,
									$fieldsIsRightTextLabelAlignment: metaData.$isRightTextLabelAlignment
								}, true);
							}
							break;
						default:
							metaData[field.$item.$bind] = value;
					}
				}
				this.designedItem.applyDesignMetaData(metaData, true);
				if (this.designedItem.$designLevel == "field") {
					for (var ii = 0, jj = _fieldAddinTypes.length; ii < jj; ii++) {
						_fieldAddinTypes[ii].notifyDataChange(this, metaData);
					}
				}
				this.history.updateSteps();
				this.applyChange(metaData);
				break;
		}
	},
	_ensureSelectedItemTitle: function() {
		this.selectedItemTitle = this.designedItem.getTitle(false);
		if (this.selectedItemTitle == "" && this.designedItem.$designLevel == "field") {
			this.selectedItemTitle = this.getBindTitle(this.designedItem.$item.$bind, this.designedItem.$field);
		}
		this.selectedPanel.itemTitle.textContent = this.selectedItemTitle && this.selectedItemTitle != "-" ? this.selectedItemTitle : "";
		this.fillTree && this.fillTree();
	},
	findItem: function(domItem, findLayout, event) {
		return syra_store.findItem(domItem);
	},
	resizeArticle: function(resize) {
		if (this.propertiesSlot) {
			var page = this.designedArticle.page;
			if (page.isVignettePage) {
				page = syra_site.landingPageMaster;
			}
			if (this.isDesignBarVisible) {
				if (page.scrollview && this.treesBar && this.treebarSlot.parentNode) {
					this.treesBar.resizeBar();
					this.asideParent.domItem.style.paddingLeft = this.treebarSlot.clientWidth + "px";
				}
				if (this.propertiesSlot.parentNode) {
					this.propertiesSlot.style.height = page.barHeight + "px";
					this.asideParent.domItem.style.paddingRight = this.propertiesSlot.clientWidth + "px";
				}
			} else {
				this.asideParent.domItem.style.paddingRight = "";
				this.asideParent.domItem.style.paddingLeft = "";
				this.designedArticle.menuBar && this.designedArticle.menuBar.resizeSplitter();
				this.designedArticle.fusionBar && this.designedArticle.fusionBar.resizeSplitter();
			}
		}
		Designer.prototype.resizeArticle.call(this, resize);
	},
	onMenuClick: function(menuItem) {
		if (this.pageStorage) {
			return this.pageStorage.onMenuClick(menuItem);
		}
		return true;
	},
	dispose: function() {
		var found = _stack.indexOf(this);
		(found >= 0) && _stack.splice(found, 1);
		_ensureDesigneBar(this, false);
		this.contentTree && this.contentTree.dispose();
		this.insertTree && this.insertTree.dispose();
		this.pageStorage && this.pageStorage.dispose();
		this.itemTool && this.itemTool.dispose();
		this.treesBar && this.treesBar.dispose();
		syra_site.dom.removeChild(this.layoutSlot);
		syra_site.dom.removeChild(this.treebarSlot);
		syra_site.dom.removeChild(this.propertiesSlot);
		syra_site.registerScreenDesigner(false, this);
		this.toggleUIDesign(this.designedArticle, false, true);
		if (this.selectedPanel) {
			this.selectedPanel.fieldAddinSection = null;
		}
		syra_site.layoutRect = null; //important for dialog resize
		Designer.prototype.dispose.call(this);
	}
});
