// syracuse-ui/lib/authoring/pageStorage.js
"use strict";
var _helpers = require('syracuse-core').helpers;
var DesktopPage = require("syracuse-ui/lib/page/desktopPage").DesktopPage;
var _datetime = require('syracuse-core').types.datetime;
var _designerViewChoice = require('syracuse-ui/lib/authoring/designerViewChoice');

function _loadSelectedDesignView(page, $selectedView) {
	var params = {
		select: "$views"
	};
	if (page.$fusionPageMeta && page.$fusionPageMeta.modelRepresentation) {
		params.modelRepresentation = page.$fusionPageMeta.modelRepresentation;
	}
	syra_ajax.get({
		page: page,
		url: syra_url.buildPageCollaborationUrl(page, page.urlSeg),
		params: params,
		success: function(data, response, requestUrl) {
			page.$views = data.$views;
			_designerViewChoice.registerDesignViews(page, $selectedView || true);
			if (!page.$views) {
				page.$item = syra_site.clone(syra_pageBuilder.ensureDefaultArticle(page, page.$prototype.$article, page.$prototype));
				page.reloadLayout(page.$item);
				if (page.urlSeg) {
					delete page.urlSeg.params.pageview;
					syra_url.build(page.urlSeg);
					if (!syra_site.mobile) {
						syra_url.history.update(page, page.urlSeg.$url);
					}
				}
			}
		}
	});
}

function PageStorage() {}

exports.PageStorage = _helpers.defineClass(PageStorage, DesktopPage, {
	loadStorage: function(designer) {
		this.isSiteRegisterDisabled = true;
		this.designer = designer;
		this.crudSlot = syra_dom.addDiv("s-aw-page-wc", designer.layoutSlot);

		this.diagnosePage = designer.designedArticle.page;
		syra_pageBuilder.initialize(this);
		if (this.designer.designedArticle == this.designer.designedArticle.page && this.designer.designedArticle.page.$authorUrl) {
			this.hasWorkingCopy = true;
			syra_form.load({
				menu: {
					$url: this.designer.designedArticle.page.$authorUrl
				},
				article: this,
				success: function() {
					syra_site.resizeItem();
				}
			});
		} else {
			this.crudSlot.syraItem = this.designer.id;
			this.designer.addCloseButton(this.crudSlot);
		}
	},
	onBeforeMainPageChange: function(continueChanging) {
		var self = this;
		if (self.designer.history.isUpdated) {
			syra_alert.ask({
				$title: syra_local.aw_updateMessageTitle,
				$message: syra_local.aw_cancelMessageText,
				$buttons: "yesnocancel",
				yes: function() {
					self.saveUpdate();
					setTimeout(function() {
						self.designer.closeDesigner();
						continueChanging();
					}, 300);
				},
				no: function() {
					self.designer.closeDesigner();
					continueChanging();
				}
			});
			return false;
		}
		return true;
	},
	onMenuClick: function(options) {
		var self = this;
		switch (options.menu.$item.$bind) {
			case "$delete":
				if (options.menu.page != self.designer.designedArticle.page) {
					if (self.designer.designedArticle.page.$selectedDesignView) {
						var $selectedDesignView = self.designer.designedArticle.page.$selectedDesignView;
						if ($selectedDesignView) {
							syra_alert.ask({
								$title: syra_local.aw_deleteMessageTitle,
								$message: syra_local.aw_deleteMessageText.replace("{0}", "'" + $selectedDesignView.$title + "'"),
								$buttons: "yesno",
								yes: function() {
									syra_router.executeMenu(options.menu);
								}
							});
						}
					}
				}
				return false;
			case "$close":
				if (self.designer.history.isUpdated) {
					syra_alert.ask({
						$title: syra_local.aw_updateMessageTitle,
						$message: syra_local.aw_cancelMessageText,
						$buttons: "yesnocancel",
						yes: function() {
							self.saveUpdate();
						},
						no: function() {
							var $item = syra_site.clone(self.designer.$sourceItem);
							self.designer.designedArticle.page.reloadLayout($item);
							syra_site.mobile && syra_site.mobile.applyChange($item);
							self.designer.closeDesigner();
						}
					});
				} else {
					self.designer.closeDesigner();
				}
				return false;
			default:
				if (options.menu.$sourceBind == "$aw_quit") {
					self.designer.clearDiagnoses();
					self.designer.closeDesigner();
					return false;
				}
				var page = options.menu.page;
				if (page != self && !page.isAlertModal && !options.menu.isDiagnoseMenu && page.isDesigned) {
					return false;
				}
		}
		return true;
	},
	saveUpdate: function() {
		syra_menus.click.fire({
			scope: this,
			$bind: (this.$menus.$save && !this.$menus.$save.$isDisabled) ? "$save" : "saveAs"
		});
	},
	drawPage: function() {
		this.$prototype.$properties = this.$prototype.$properties || {};
		this.addItem(this.crudSlot, {
			$bind: "$save",
			$category: "link",
			$skin: "s-aw-menus-save",
			$icon: {
				$mode: "icon"
			}
		});
		if (!syra_site.mobile) {
			this.addItem(this.crudSlot, {
				$bind: "saveAs",
				$category: "link",
				$skin: "s-aw-menus-saveas",
				$icon: {
					$mode: "icon",
					$path: "authoring/s-aw-"
				}
			});
			this.addItem(this.crudSlot, {
				$bind: "$delete",
				$category: "link",
				$skin: "s-aw-menus-delete",
				$icon: {
					$mode: "icon"
				}
			});
		}
		this.addItem(this.crudSlot, {
			$bind: "$close",
			$category: "link",
			$skin: "s-aw-page-close",
			$icon: {
				$mode: "icon"
			}
		});
		this.menuItems.$close[0].setMenu({
			$title: syra_local.aw_close
		}, null);
	},

	onAfterActionMenuExecute: function(menuItem, $menu, isSuccess) {
		switch (menuItem.$bind) {
			case "$save":
			case "saveAs":
				if (isSuccess) {
					if (this.dataset.$authorUrl) {
						this.designer.designedArticle.page.$authorUrl = this.dataset.$authorUrl;
					}
					this.designer.history.updateSteps(false);
					$menu.$diagnoses = [{
						$message: syra_local.aw_saveMessageText,
						$severity: "success"
					}];
					$menu.$links = {
						$create: {
							$isHidden: true
						},
						$details: {
							$isHidden: true
						},
						$query: {
							$isHidden: true
						}
						/*,
                     $aw_quit: {
                     $title: syra_local.aw_quit
                     }*/
					};
					syra_dataset.applyDelta(menuItem.page, menuItem.articleParent.$menus[menuItem.$bind], {
						$links: $menu.$links
					});
					menuItem.disable(false);
					if (menuItem.$bind == "saveAs") {
						_loadSelectedDesignView(this.designer.designedArticle.page, menuItem.$links && menuItem.$links.$view);
					}
				} else {
					if (menuItem.$bind == "saveAs") {
						menuItem.disable(false);
					}
				}
				break;
			case "$delete":
				if (isSuccess) {
					var $selectedDesignView = this.designer.designedArticle.page.$selectedDesignView;
					if (this.dataset.$authorUrl) {
						this.designer.designedArticle.page.$authorUrl = this.dataset.$authorUrl;
					}
					var $diagnose = {
						$message: syra_local.aw_isDeletedMessageText.replace("{0}", "'" + $selectedDesignView.$title + "'"),
						$severity: "success"
					};
					var page = this.designer.designedArticle.page;
					this.designer.closeDesigner();
					_loadSelectedDesignView(page);
					return false;
				}
				break;
		}
	},
	applyChange: function(newData) {
		if (newData) {
			newData.$diagnoses = null;
		}
		DesktopPage.prototype.applyChange.call(this, newData);
	},
	form_postAction: function(menuItem, target, value) {
		if (value && value.$save || value.saveAs) {
			if (this.designer.designedArticle.page.garbage) {
				this.designer.designedArticle.page.garbage.garbageFreeItems();
			}
		}
		var $article = syra_site.clone(this.designer.designedArticle.page.$item);
		delete $article.$menus;
		delete $article.$isModel;
		$article.$updateDate = _datetime.now().toString("yyyy-MM-ddTHH:mm:ss") + "Z";
		if (this.designer.designedArticle.page.makeArticlePersistent) {
			this.designer.designedArticle.page.makeArticlePersistent($article);
		}
		this.saveDesign($article);
		return true;
	},
	saveDesign: function($article) {
		if (this.hasWorkingCopy) {
			syra_form.getSendBag(this).content = {
				$article: syra_layoutUpdater.cleanBeforeSave($article || this.designer.designedArticle.page.$item)
			};
		}
	},
	dispose: function() {
		syra_dom.remove(this.crudSlot);
		DesktopPage.prototype.dispose.call(this);
	}
});
