// syracuse-ui/lib/authoring/paletteFieldsProvider.js
"use strict";
var authoringTypes = {
	choiceField: require("syracuse-ui/lib/authoring/fieldPalette/choiceFieldPalette"),
	documentField: require("syracuse-ui/lib/authoring/fieldPalette/documentFieldPalette")
};

exports.addFieldProperties = function($properties, localize) {
	authoringTypes.choiceField.addProperties($properties, localize);
	authoringTypes.documentField.addProperties($properties, localize);
};

exports.addFieldSection = function($layout) {
	if ($layout.$items.length == 0) {
		$layout.$items.push(authoringTypes.choiceField.getSection());
		$layout.$items.push(authoringTypes.documentField.getSection());
	}
};

exports.onShow = function(awPalette, awItem) {
	authoringTypes.choiceField.onShow(awPalette, awItem);
	authoringTypes.documentField.onShow(awPalette, awItem);
};

exports.notifyDataChange = function(awPalette, awItem, metaData) {
	authoringTypes.choiceField.notifyDataChange(awPalette, awItem, metaData);
	authoringTypes.documentField.notifyDataChange(awPalette, awItem, metaData);
};

exports.applyChange = function(awPalette, awItem, newData) {
	authoringTypes.choiceField.applyChange(awPalette, awItem, newData);
	authoringTypes.documentField.applyChange(awPalette, awItem, newData);
};
