// syracuse-ui/lib/authoring/paletteSectionProvider.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function _getLayoutSection($level, localize) {
	var $layout = {
		$category: "section",
		$skin: "s-aw-h2-" + $level,
		$isBoxCollapsable: true,
		$isColumnSection: true,
		$columnSectionCss: $level == "field" ? "" : "s-aw-row-box",
		$isStructure: true,
		$title: localize.aw_layout,
		$layout: {
			$items: [{
				$category: "field",
				$isHidden: !($level == "article" || $level == "section"),
				$css: "s-layout-tab-choice",
				$bind: "$layoutTabChoice",
				$skin: "s-aw-field",
				$isEditMode: true,
				$format: "$button",
				$choiceLayout: "row",
				$icon: {
					$mode: "icon",
					$path: "authoring/s-layout-"
				}
			}]
		}
	};
	return $layout;
}

function _getAppearanceSection($level, localize) {
	return {
		$category: "section",
		$skin: "s-aw-h2-" + $level,
		$isBoxCollapsable: true,
		$isStructure: true,
		$title: localize.aw_appearance,
		$layout: {
			$items: [{
				$category: "field",
				$css: "s-layout-tab-choice",
				$bind: "$boxTabChoice",
				$skin: "s-aw-field",
				$isEditMode: true,
				$format: "$button",
				$choiceLayout: "row",
				$icon: {
					$mode: "icon",
					$path: "authoring/s-layout-"
				}
			}]
		}
	};
}

function _getRowSection(localize) {
	return {
		$category: "section",
		$skin: "s-aw-h2-layout",
		$isRowSection: true,
		$isColumnSection: true,
		$title: localize.aw_rowLayout,
		$layout: {
			$items: [{
				$category: "field",
				$bind: "$rowAlign",
				$skin: "s-aw-field",
				$isEditMode: true
			}]
		}
	};
}

exports.fillColChoices = function(localize, section, colCount) {
	section.dataStore = section.dataStore || {};
	var css = "s-aw-cols-link " + (section.$item.$columnSectionCss || "");
	if (!section.dataStore.colChoices) {
		section.dataStore.colChoices = {};
		var row = document.createElement("div");
		row.className = "s-aw-cols";
		var cell = document.createElement("div");
		cell.className = "s-aw-cols-label";
		cell.textContent = localize.aw_columns;
		row.appendChild(cell);

		for (var ii = 1; ii <= 6; ii++) {
			var link = document.createElement("a");
			link.textContent = link.$syraCoCount = ii;
			link.className = css;
			link.setAttribute("data-s-picker", "$colcount");
			row.appendChild(section.dataStore.colChoices[ii] = link);
		}
		var nextSibling = section.dataStore.insertRow ? section.dataStore.insertRow.nextSibling : section._body.firstChild;
		nextSibling.parentNode.insertBefore(row, nextSibling);
	}
	for (var ii = 1; ii <= 6; ii++) {
		section.dataStore.colChoices[ii].className = css + (colCount == ii ? " s-selected" : "");
	}
};
exports.fillRowSection = function(localize, section) {
	section.dataStore = section.dataStore || {};
	if (!section.dataStore.insertRow) {
		var row = document.createElement("div");
		row.className = "s-aw-insert-row";

		var cell = document.createElement("div");
		cell.className = "s-aw-insert-row-label";
		cell.textContent = localize.aw_addrow;
		row.appendChild(cell);

		cell = document.createElement("div");
		cell.className = "s-aw-insert-row-links";
		var links = ["addRowBefore", "addRowAfter"];
		for (var ii = 0, jj = links.length; ii < jj; ii++) {
			var link = document.createElement("a");
			link.textContent = localize["aw_" + links[ii]];
			link.className = "s-aw-insert-row-link";
			link.setAttribute("data-s-picker", "$" + links[ii]);
			cell.appendChild(link);
		}
		row.appendChild(cell);
		section._body.insertBefore(section.dataStore.insertRow = row, section._body.firstChild);
	}
};

exports.article = function(localize) {
	return {
		$category: "section",
		$skin: "s-aw-h1-article",
		$title: localize.aw_page,
		$isHidden: true,
		$layout: {
			$items: [{
				$category: "section",
				$isTitleHidden: true,
				$skin: "s-aw-palette-item",
				$layout: {
					$layoutType: "stack",
					$items: [_getRowSection(localize), _getLayoutSection("article", localize), {
						$category: "section",
						$skin: "s-aw-h2-article",
						$bind: "$field",
						$isSectionFields: true,
						$isBoxCollapsable: true,
						$title: localize.aw_sectionFields,
						$layout: {
							$layoutType: "stack",
							$items: [{
								$bind: "$fieldsIsTitleHidden",
								$skin: "s-aw-field",
								$isEditMode: true,
								$isLeftSpaceHidden: true
							}, {
								$bind: "$labelAlignment",
								$skin: "s-aw-field",
								$isEditMode: true,
								$isLeftSpaceHidden: true,
								$choiceLayout: "1"
							}]
						}
					}, {
						$category: "section",
						$skin: "s-aw-h2-article",
						$isBoxCollapsable: true,
						$isSectionModels: true,
						$title: localize.aw_models
					}]
				}
			}]
		}
	};
};


exports.section = function(localize) {
	return {
		$category: "section",
		$skin: "s-aw-h1-section",
		$title: localize.aw_section,
		$isHidden: true,
		$layout: {
			$items: [{
				$category: "section",
				$isTitleHidden: true,
				$skin: "s-aw-palette-item",
				$layout: {
					$layoutType: "stack",
					$items: [{
							$bind: "$excludeMe",
							$category: "link",
							$skin: "s-aw-menus-link",
							$css: "s-aw-exclude-me"
						},
						_getRowSection(localize), _getAppearanceSection("section", localize), _getLayoutSection("section", localize), {
							$category: "section",
							$skin: "s-aw-h2-section",
							$isBoxCollapsable: true,
							$isTitleSection: true,
							$title: localize.aw_sectionAppearance,
							$layout: {
								$items: [{
									$bind: "$title",
									$skin: "s-aw-field",
									$isEditMode: true,
									$isLeftSpaceHidden: true
								}, {
									$bind: "$titleChoice",
									$skin: "s-aw-field",
									$isEditMode: true,
									$choiceLayout: "1"
								}, {
									$bind: "$titleLocalization",
									$skin: "s-aw-field",
									$isEditMode: true,
									$format: "$combo"
								}]
							}
						}, {
							$category: "section",
							$skin: "s-aw-h2-section",
							$isBoxCollapsable: true,
							$title: localize.aw_sectionBehaviour,
							$layout: {
								$items: [{
									$bind: "$isBoxCollapsable",
									$skin: "s-aw-field",
									$isEditMode: true,
									$isLeftSpaceHidden: true
								}, {
									$bind: "$isMaximizable",
									$skin: "s-aw-field",
									$isEditMode: true,
									$isLeftSpaceHidden: true
								}]
							}
						}, {
							$category: "section",
							$skin: "s-aw-h2-section",
							$isBoxCollapsable: true,
							$bind: "$field",
							$isSectionFields: true,
							$title: localize.aw_sectionFields,
							$layout: {
								$layoutType: "stack",
								$items: [{
									$bind: "$fieldsIsTitleHidden",
									$skin: "s-aw-field",
									$isEditMode: true,
									$isLeftSpaceHidden: true
								}, {
									$bind: "$labelAlignment",
									$skin: "s-aw-field",
									$isEditMode: true,
									$isLeftSpaceHidden: true,
									$choiceLayout: "1"
								}]
							}
						}, {
							$category: "section",
							$skin: "s-aw-h2-section",
							$isBoxCollapsable: true,
							$isSectionModels: true,
							$title: localize.aw_models
						}
					]
				}
			}]
		}
	};
};

exports.block = function(localize) {
	return {
		$category: "section",
		$skin: "s-aw-h1-block",
		$title: localize.aw_block,
		$isHidden: true,
		$layout: {
			$items: [{
				$category: "section",
				$isTitleHidden: true,
				$skin: "s-aw-palette-item",
				$layout: {
					$layoutType: "stack",
					$items: [{
							$bind: "$excludeMe",
							$category: "link",
							$skin: "s-aw-menus-link",
							$css: "s-aw-exclude-me"
						},
						_getRowSection(localize), _getAppearanceSection("block", localize), _getLayoutSection("block", localize), {
							$category: "section",
							$skin: "s-aw-h2-block",
							$isBoxCollapsable: true,
							$isTitleSection: true,
							$title: localize.aw_sectionAppearance,
							$layout: {
								$items: [{
									$bind: "$title",
									$skin: "s-aw-field",
									$isEditMode: true,
									$isLeftSpaceHidden: true
								}, {
									$bind: "$titleChoice",
									$skin: "s-aw-field",
									$isEditMode: true,
									$choiceLayout: "1"
								}, {
									$bind: "$titleLocalization",
									$skin: "s-aw-field",
									$isEditMode: true,
									$format: "$combo"
								}]
							}
						}, {
							$category: "section",
							$skin: "s-aw-h2-block",
							$isBoxCollapsable: true,
							$title: localize.aw_sectionBehaviour,
							$layout: {
								$items: [{
									$bind: "$isBoxCollapsable",
									$skin: "s-aw-field",
									$isEditMode: true,
									$isLeftSpaceHidden: true
								}, {
									$bind: "$isMaximizable",
									$skin: "s-aw-field",
									$isEditMode: true,
									$isLeftSpaceHidden: true
								}]
							}
						}, {
							$category: "section",
							$skin: "s-aw-h2-block",
							$isBoxCollapsable: true,
							$bind: "$field",
							$isSectionFields: true,
							$title: localize.aw_sectionFields,
							$layout: {
								$layoutType: "stack",
								$items: [{
									$bind: "$fieldsIsTitleHidden",
									$skin: "s-aw-field",
									$isEditMode: true,
									$isLeftSpaceHidden: true
								}, {
									$bind: "$labelAlignment",
									$skin: "s-aw-field",
									$isEditMode: true,
									$isLeftSpaceHidden: true,
									$choiceLayout: "1"
								}]
							}
						}, {
							$category: "section",
							$skin: "s-aw-h2-block",
							$isBoxCollapsable: true,
							$isSectionModels: true,
							$title: localize.aw_models
						}
					]
				}
			}]
		}
	};
};

exports.field = function(localize) {
	return {
		$category: "section",
		$skin: "s-aw-h1-field",
		$title: localize.aw_field,
		$isHidden: true,
		$layout: {
			$items: [{
				$category: "section",
				$isTitleHidden: true,
				$id: "s-aw-palette-field",
				$skin: "s-aw-palette-item",
				$layout: {
					$layoutType: "stack",
					$items: [{
							$bind: "$excludeMe",
							$category: "link",
							$skin: "s-aw-menus-link",
							$css: "s-aw-exclude-me"
						}, {
							$bind: "$isAdvanced",
							$skin: "s-aw-field",
							$isEditMode: true,
							$isLeftSpaceHidden: true
						},
						_getRowSection(localize), _getLayoutSection("field", localize), {
							$category: "section",
							$skin: "s-aw-h2-field",
							$isLabelSection: true,
							$isBoxCollapsable: true,
							$title: localize.aw_sectionFieldLabel,
							$layout: {
								$layoutType: "stack",
								$items: [{
									$bind: "$isTitleHidden",
									$title: localize.aw_isLabelHidden,
									$skin: "s-aw-field",
									$isEditMode: true,
									$isLeftSpaceHidden: true
								}, {
									$bind: "$labelAlignment",
									$title: localize.aw_sectionFieldLabelPosition,
									$skin: "s-aw-field",
									$isEditMode: true,
									$choiceLayout: "1"
								}]
							}
						}, {
							$category: "section",
							$title: localize.aw_sectionFieldWidget,
							$isBoxCollapsable: true,
							$skin: "s-aw-h2-field",
							$isHidden: true,
							$isFieldAddinSection: true,
							$layout: {
								$items: []
							}
						}
					]
				}
			}]
		}
	};
};

exports.menu = function(localize) {
	return {
		$category: "section",
		$skin: "s-aw-h1-field",
		$title: localize.aw_menuItem,
		$isHidden: true,
		$layout: {
			$items: [{
				$category: "section",
				$isTitleHidden: true,
				$skin: "s-aw-palette-item",
				$layout: {
					$layoutType: "stack",
					$items: [{
							$bind: "$excludeMe",
							$category: "link",
							$skin: "s-aw-menus-link",
							$css: "s-aw-exclude-me"
						},
						_getRowSection(), _getLayoutSection("field", localize)
					]
				}
			}]
		}
	};
};
