// syracuse-ui/lib/authoring/sectionProvider.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function _getLayoutSection($level, localize) {
	var $layout = {
		$category: "section",
		$skin: "s-aw-h2-" + $level,
		$isSectionModels: true,
		$isStructure: true,
		$title: localize.aw_template,
		$layout: {
			$items: []
		}
	};
	return $layout;
}

function _getAppearanceSection($level, localize) {
	return {
		$category: "section",
		$skin: "s-aw-h2-" + $level,
		$isTabChoice: "$boxTabChoice",
		$isStructure: true,
		$title: localize.aw_appearance,
		$layout: {
			$items: []
		}
	};
}

function _getRowSection(localize) {
	return {
		$category: "section",
		$skin: "s-aw-h2-layout",
		$isRowSection: true,
		$isSectionModels: true,
		$isColumnSection: true,
		$title: localize.aw_rowLayout,
		$layout: {
			$items: [{
				$category: "field",
				$bind: "$rowAlign",
				$skin: "s-aw-field",
				$isEditMode: true
			}]
		}
	};
}



exports.fillRowSection = function(localize, section) {
	section.dataStore = section.dataStore || {};
	if (!section.dataStore.insertRow) {
		var row = document.createElement("div");
		row.className = "s-aw-insert-row";

		var cell = document.createElement("div");
		cell.className = "s-aw-insert-row-label";
		cell.textContent = localize.aw_addrow;
		row.appendChild(cell);

		cell = document.createElement("div");
		cell.className = "s-aw-insert-row-links";
		var links = ["addRowBefore", "addRowAfter"];
		for (var ii = 0, jj = links.length; ii < jj; ii++) {
			var link = document.createElement("a");
			link.textContent = localize["aw_" + links[ii]];
			link.className = "s-aw-insert-row-link";
			link.setAttribute("data-s-picker", "$" + links[ii]);
			cell.appendChild(link);
		}
		row.appendChild(cell);
		section._body.insertBefore(section.dataStore.insertRow = row, section._body.firstChild);
	}
};

exports.article = function(localize) {
	return {
		$category: "section",
		$skin: "s-aw-h1-article",
		$title: localize.aw_page,
		$isHidden: true,
		$layout: {
			$layoutType: "stack",
			$items: [_getRowSection(localize), _getLayoutSection("article", localize), {
				$category: "section",
				$skin: "s-aw-h2-article",
				$bind: "$field",
				$isSectionFields: true,
				$title: localize.aw_sectionFields,
				$layout: {
					$layoutType: "stack",
					$items: [{
						$bind: "$fieldsIsTitleHidden",
						$skin: "s-aw-field",
						$isEditMode: true,
						$isLeftSpaceHidden: true
					}, {
						$bind: "$labelAlignment",
						$skin: "s-aw-field",
						$isEditMode: true,
						$isLeftSpaceHidden: true,
						$choiceLayout: "1"
					}]
				}
			}]
		}
	};
};


exports.section = function(localize) {
	return {
		$category: "section",
		$skin: "s-aw-h1-section",
		$title: localize.aw_section,
		$isHidden: true,
		$layout: {
			$layoutType: "stack",
			$items: [{
					$bind: "$excludeMe",
					$category: "link",
					$skin: "s-aw-exclude-me"
				},
				_getRowSection(localize), _getAppearanceSection("section", localize), _getLayoutSection("section", localize), {
					$category: "section",
					$skin: "s-aw-h2-section",
					$isTitleSection: true,
					$title: localize.aw_sectionAppearance,
					$layout: {
						$items: [{
							$bind: "$title",
							$skin: "s-aw-field",
							$isEditMode: true,
							$isLeftSpaceHidden: true
						}, {
							$bind: "$titleChoice",
							$skin: "s-aw-field",
							$isEditMode: true,
							$choiceLayout: "1"
						}, {
							$bind: "$titleLocalization",
							$skin: "s-aw-field",
							$isEditMode: true,
							$format: "$combo"
						}]
					}
				}, {
					$category: "section",
					$skin: "s-aw-h2-section",
					$title: localize.aw_sectionBehaviour,
					$layout: {
						$items: [{
							$bind: "$isBoxCollapsable",
							$skin: "s-aw-field",
							$isEditMode: true,
							$isLeftSpaceHidden: true
						}, {
							$bind: "$isMaximizable",
							$skin: "s-aw-field",
							$isEditMode: true,
							$isLeftSpaceHidden: true
						}]
					}
				}, {
					$category: "section",
					$skin: "s-aw-h2-section",
					$bind: "$field",
					$isSectionFields: true,
					$title: localize.aw_sectionFields,
					$layout: {
						$layoutType: "stack",
						$items: [{
							$bind: "$fieldsIsTitleHidden",
							$skin: "s-aw-field",
							$isEditMode: true,
							$isLeftSpaceHidden: true
						}, {
							$bind: "$labelAlignment",
							$skin: "s-aw-field",
							$isEditMode: true,
							$isLeftSpaceHidden: true,
							$choiceLayout: "1"
						}]
					}
				}
			]
		}
	};
};

exports.block = function(localize) {
	return {
		$category: "section",
		$skin: "s-aw-h1-block",
		$title: localize.aw_block,
		$isHidden: true,
		$layout: {
			$layoutType: "stack",
			$items: [{
					$bind: "$excludeMe",
					$category: "link",
					$skin: "s-aw-exclude-me"
				},
				_getRowSection(localize), _getAppearanceSection("block", localize), _getLayoutSection("block", localize), {
					$category: "section",
					$skin: "s-aw-h2-block",
					$isTitleSection: true,
					$title: localize.aw_sectionAppearance,
					$layout: {
						$items: [{
							$bind: "$title",
							$skin: "s-aw-field",
							$isEditMode: true,
							$isLeftSpaceHidden: true
						}, {
							$bind: "$titleChoice",
							$skin: "s-aw-field",
							$isEditMode: true,
							$choiceLayout: "1"
						}, {
							$bind: "$titleLocalization",
							$skin: "s-aw-field",
							$isEditMode: true,
							$format: "$combo"
						}]
					}
				}, {
					$category: "section",
					$skin: "s-aw-h2-block",
					$title: localize.aw_sectionBehaviour,
					$layout: {
						$items: [{
							$bind: "$isBoxCollapsable",
							$skin: "s-aw-field",
							$isEditMode: true,
							$isLeftSpaceHidden: true
						}, {
							$bind: "$isMaximizable",
							$skin: "s-aw-field",
							$isEditMode: true,
							$isLeftSpaceHidden: true
						}]
					}
				}, {
					$category: "section",
					$skin: "s-aw-h2-block",
					$bind: "$field",
					$isSectionFields: true,
					$title: localize.aw_sectionFields,
					$layout: {
						$layoutType: "stack",
						$items: [{
							$bind: "$fieldsIsTitleHidden",
							$skin: "s-aw-field",
							$isEditMode: true,
							$isLeftSpaceHidden: true
						}, {
							$bind: "$labelAlignment",
							$skin: "s-aw-field",
							$isEditMode: true,
							$isLeftSpaceHidden: true,
							$choiceLayout: "1"
						}]
					}
				}
			]
		}
	};
};

exports.field = function(localize) {
	return {
		$category: "section",
		$skin: "s-aw-h1-field",
		$title: localize.aw_field,
		$isHidden: true,
		$layout: {
			$layoutType: "stack",
			$items: [{
					$bind: "$excludeMe",
					$category: "link",
					$skin: "s-aw-exclude-me"
				}, {
					$bind: "$isAdvanced",
					$skin: "s-aw-field",
					$isEditMode: true,
					$isLeftSpaceHidden: true
				},
				_getRowSection(localize), _getLayoutSection("field", localize), {
					$category: "section",
					$skin: "s-aw-h2-field",
					$isLabelSection: true,
					$title: localize.aw_sectionFieldLabel,
					$layout: {
						$layoutType: "stack",
						$items: [{
							$bind: "$isTitleHidden",
							$title: localize.aw_isLabelHidden,
							$skin: "s-aw-field",
							$isEditMode: true,
							$isLeftSpaceHidden: true
						}, {
							$bind: "$labelAlignment",
							$title: localize.aw_sectionFieldLabelPosition,
							$skin: "s-aw-field",
							$isEditMode: true,
							$choiceLayout: "1"
						}]
					}
				}, {
					$category: "section",
					$title: localize.aw_sectionFieldWidget,
					$skin: "s-aw-h2-field",
					$isHidden: true,
					$isFieldAddinSection: true,
					$layout: {
						$items: []
					}
				}
			]
		}
	};
};
