// syracuse-ui/lib/page/landing/search/landingSearchVignettePage.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var DesktopPage = require('syracuse-ui/lib/page/desktopPage').DesktopPage;
var VignettePage = require('syracuse-ui/lib/page/landing/vignette/vignettePage').VignettePage;

function LandingSearchVignettePage() {}

exports.LandingSearchVignettePage = helpers.defineClass(LandingSearchVignettePage, VignettePage, {
	_ensureMenus: function() {
		VignettePage.prototype._ensureMenus.call(this);
		DesktopPage.prototype._ensureMenus.call(this);
		this.menuBar.barSlot &&
			this.loadNewItem(this.menuBar.barSlot, {
				$category: "landingVignetteSearchFacets",
				$prototype: {},
				$bind: "$searchFacets",
				$isVignetteSearchFacet: true,
				$skin: "s-sv-search-facets",
				$itemSkin: "s-sv-search-facet"
			});
	},
	fetch: function(options, callback, error) {
		this.vignetteField && this.vignetteField.page.fetch(options);
	},
	onMenuClick: function(menu) {
		return this.vignetteField.page.onMenuClick(menu);
	},
	onSelectRecord: function(records) {
		if (records) {
			var keys = Object.keys(records);
			keys.length && syra_menus.click.detail(records[keys[0]]);
		}
	},
	updateFilterState: function(options) {
		this.vignetteField.page.updateFilterState(options);
	}
});
