// syracuse-ui/lib/common/article/menuItem.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function _showDiagnoses(menu, $delta){
    var article = menu.getArticle();
    var $menu = article.$menus[menu.$item.$bind];
    if ($menu && $menu.$links) {
        var $links = $menu.$links || {};
        if ($delta.$links) {
            Object.keys($delta.$links).forEach(function($bind){
                var $source = $delta.$links[$bind];
                var $target = $links[$bind] = $links[$bind] || {};
                Object.keys($source).forEach(function($prop){
                    $target[$prop] = $source[$prop];
                });
            });
        }
        Object.keys($links).forEach(function($bind){
            var $link = $links[$bind];
            $link.$url = article.parseExpression($link.$url);
        });
        document.site.showDiagnoses({
            $links: $links,
            $diagnoses: $delta.$diagnoses
        }, article);
    }
}

function _onMenuItemClick(menuItem){
    var doClick = true;
    if (menuItem.boxParent) {
        doClick = menuItem.boxParent.onMenuClick ? menuItem.boxParent.onMenuClick(menuItem) : true;
        var article = menuItem.getArticle();
        doClick = article.onMenuClick ? article.onMenuClick(menuItem) : true;
    }
    if (doClick) {
        if (menuItem.onMenuClick ? menuItem.onMenuClick() : true) {
            document.controller.executeMenu(menuItem);
        }
    }
}

function MenuItem(){

}

exports.MenuItem = helpers.defineClass(MenuItem, null, {
    loadBox: function(record){
        this.$skin = this.$item.$skin || "s-link";
        if (this.$item.$bind) {
            this.getArticle().bindMenuItem(this, record);
        }
        else 
            if (this.$item.$link) {
                this.setMenu(this.$item.$link);
            }
    },
    getArticle: function(){
        if (!this.article) {
            this.article = this.boxParent ? this.boxParent.getArticle() : null;
        }
        return this.article;
    },
    getPage: function(){
        if (!this.page) {
            this.page = this.getArticle().getPage();
        }
        return this.page;
    },
    setTitle: function($value){
        this.$title = $value;
        if (this.$item.$noText) {
            this.domItem.title = $value;
        }
        else {
            if ($value.indexOf("{") < 0) {
                this.$$item.text($value);
            }
            else {
                this.boxParent._renderExpression($value, this.$$item);
            }
        }
    },
    setDescription: function($value){
        this.$description = $value;
        if (this.$$description) {
            this.$$description.text($value);
        }
        else {
            this.domItem.title = $value;
        }
    },
    hide: function($value){
        this.$isHidden = $value;
        this.domItem.style.display = !$value ? "" : "none"
        if (this.$$liParent) {
            this.$$liParent[0].style.display = !$value ? "" : "none";
        }
    },
    disable: function($value){
        this.$isDisabled = $value || false;
        this.$$item.attr("disabled", this.$isDisabled).toggleClass("s-disabled", this.$isDisabled);
    },
    
    click: function(){
        var self = this;
        if (!self.$isDisabled) {
            var page = self.getPage();
            if (page) {
                page.externalAdapter.onMenuItemClick({
                    menuItem: self,
                    doEvent: function(){
                        _onMenuItemClick(self);
                    }
                });
            }
            else {
                _onMenuItemClick(self);
            }
        }
    },
    _createMenu: function(article, $menu){
        this.$$item = $(this.domItem = document.createElement("a"));
        this.domItem.className = this.$skin + ((this.$item.$css) ? (" " + this.$item.$css) : "");
        this.domItem.setAttribute("data-s-menu", this.id);
        this.domItem.setAttribute("data-s-article", article.id);
        this.$$item.appendTo(this.$$container);
        if (this.$item.$isDescriptionVisible) {
            // if (this.$item.$isDescriptionInline) {
            var label = document.createElement("a");
            label.className = this.$item.$skinDescription || (this.$skin + "-desc");
            this.$$description = $(label).appendTo(this.$$container);
            //}
        }
        if (this.$item.$isHidden) {
            this.hide(true);
        }
        if (!$menu.$title && this.$item.$title) {
            $menu.$title = this.$item.$title; //use for pager (next prev) currently
        }
    },
    setMenu: function($menu, record){
        var self = this;
        var article = self.getArticle();
        if ($menu) {
            if (!self.$$item) {
                self._createMenu(article, $menu);
            }
            if ($menu.$title !== undefined) {
                self.setTitle($menu.$title);
            }
            if ($menu.$description !== undefined) {
                self.setDescription($menu.$description);
            }
            if ($menu.$isHidden !== undefined) {
                self.hide($menu.$isHidden);
            }
            if ($menu.$isDisabled !== undefined) {
                self.disable($menu.$isDisabled);
            }
            Object.keys($menu).forEach(function($property){
                self[$property] = $menu[$property];
            });
            if (self.$isDisabled && $menu.$isRequested === false) {
                if (self.$item.onServerRequest ? self.$item.onServerRequest(self) : true) {
                    if ($menu.$diagnoses) {
                        _showDiagnoses(self, $menu);
                    }
                    if ($menu.$clientAgent) {
                        var agent = document.site.agents[$menu.$clientAgent.$id];
                        if (agent) {
                            agent[$menu.$clientAgent.$action](self, $menu, record);
                        }
                    }
                    if ($menu.$links && $menu.$links.$location) {
                        var $url = $menu.$links.$location.$url;
                        setTimeout(function(){
                            document.site.unload();
                            window.open($url, "_self");
                        }, 100);
                    }
                }
                $menu.$isRequested = true; //avoid procsessing action for other menu bound to $menu (create menu for sample)
            }
        }
        if (self.$type && self.$type.indexOf('json') == -1) {
            if (self.$url && (self.$url.indexOf("format=") < 0)) { // TODO: improve this hack
                self.$url += "&format=" + self.formatUrl(record, self.$type);
            }
            self.domItem.setAttribute("target", self.$target = "blank");
            self.domItem.setAttribute("href", self.$url = self.formatUrl(record));
            self.domItem.className += " " + self.$skin + "-" + self.$type.replace("application/", "");
        }
        else {
            if (self.domItem) {
                if (self.$url) {
                    self.domItem.setAttribute("href", "?url=" + encodeURIComponent(self.$url = self.formatUrl(record)));
                }
                else {
                    self.domItem.setAttribute("href", "#");
                }
            }
        }
    },
    parseProperties: function(record){
        if (this.$bind == "employeeProperties") {
            debugger;
        }
        if (this.$properties) {
            var self = this;
            var values = {};
            Object.keys(self.$properties).forEach(function(prop){
                values[prop] = self.article.parseExpression(self.$properties[prop], record);
            });
            return values;
        }
        return null;
    },
    formatUrl: function(record, $url){
        return this.article.parseExpression($url || this.$url, record, this.parseProperties(record));
    },
    dispose: function(){
        if (this.article) {
            this.article.unbindMenuItem(this, false);
            delete this.article;
        }
        delete this.page;
        delete this.onMenuClick;
    }
});
