// syracuse-ui/lib/common/article/page.js
"use strict";
//http://teamtreehouse.com/
var helpers = require('syracuse-core/lib/helpers');
var Article = require('./article').Article;

var _$menus = {
    $search: {
        $linkBox: {
            $excludeBind: ["$create", "$details", "$edit", "$prototype", "$first", "$previous", "$next", "$last"]
        }
    },
    $cube: {
        $mainAction: "$create",
        $linkBox: {
            $excludeBind: ["$details", "$edit", "$prototype", "$first", "$previous", "$next", "$last"]
        }
    },
    $query: {
        $mainAction: "$create",
        $linkBox: {
            $excludeBind: ["$details", "$edit", "$prototype", "$first", "$previous", "$next", "$last"]
        }
    },
    $details: {
        $mainAction: "$edit",
        $linkBox: {
            $excludeBind: ["$draft", "$stylesheet"]
        }
    },
    $edit: {
        $mainAction: "$save",
        $linkBox: {
            $excludeBind: ["$draft", "$stylesheet", "$prototype"]
        }
    }
};

/*****************************************************************************************************************
 **************************************    Basic Page	(CRUD, lookup)  ******************************************
 ******************************************************************************************************************/
function Page(){
}

exports.Page = helpers.defineClass(Page, Article, {
    appendOkButton: function(){
        this._appendButton("s-dialog-page-ok", "onValidate", true);
    },
    
    _appendButton: function($skin, action, first){
        var self = this;
        self.$$header[first ? "prepend" : "append"]($("<a/>").addClass($skin).bind("click", function(){
            if (self[action] ? self[action](self) : true) {
                if (self.boxParent) {
                    self.getArticleParent().removeItem(self, true);
                }
                else {
                    self.dispose();
                }
            }
        }));
    },
    _getFacetMenus: function(){
        return _$menus[this.$facet];
    },
    appendArticleMenus: function(){
        var $menus = this._getFacetMenus();
        if ($menus) {
            this.$$menus = $("<div/>").addClass(this.$skin + "-menus").appendTo(this.$$item);
            document.itemFactory.load(this.$$menus, {
                $mainAction: $menus.$mainAction,
                $bind: "$linkBox",
                $skin: "s-h1-menus",
                $category: "links",
                $excludeBind: $menus.$linkBox.$excludeBind
            }, this);
        }
    },
    getPage: function(){
        return this;
    },
    
    loadBox: function(initData){
        this.$authoringLevel = "article";
        this.externalAdapter = this.$item.externalAdapter || document.site.externalAdapter;
        delete this.$item.externalAdapter;
        Article.prototype.loadBox.call(this, initData);
        if (this.$autoFetch = this.$autoFetch !== false) {
            this.fetch();
        }
    },
    fetch: function(options, callback){
        document.controller.sendRequest(this, options, callback);
    }
});
