// syracuse-ui/lib/common/tests/arrayFieldsTest.js
"use strict";
var helpers = require("syracuse-core/lib/helpers")
var FieldsTest = require("./fieldsTest").FieldsTest;

function ArrayFieldsTest(){
}

exports.ArrayFieldsTest = helpers.defineClass(ArrayFieldsTest, FieldsTest, {
    onNotifyRecordChange: function(value, binding){
        var self = this;
        if (binding == "$actions") {
            if (value.$deleteItem || value.$insertItem || value.$reorderItem) {
                var fieldServerData = self.boundFields.$fieldServerData[0];
                fieldServerData.setDataValue("");
                Object.keys(self.boundFields).forEach(function($bind){
                    if ($bind.indexOf("array") == 0) {
                        self.boundFields[$bind].forEach(function(field, index){
                            if (index == 0) {
                                var data = {};
                                data[$bind] = field.builder._records.map(function(row){
                                    return row.boundFields.$singleField[0].getDataValue();
                                });
                                if (value.$deleteItem) {
                                    data[$bind].shift();
                                }
                                if (value.$insertItem) {
                                    var newValue;
                                    switch ($bind) {
                                        case "arrayString":
                                            newValue = "new value";
                                            break;
                                        case "arrayEmail":
                                            newValue = "poupou@pidou.com";
                                            break;
                                        case "arrayChoice":
                                            newValue = 2;
                                            break;
                                        case "arrayDate":
                                            newValue = "2012-01-15"
                                            break;
                                        case "arrayBoolean":
                                            newValue = true;
                                            break;
                                        case "arrayDecimal":
                                            newValue = 454511.25;
                                            break;
                                        case "arrayDecimal":
                                            newValue = 454511.25;
                                            break;
                                        case "arrayReference":
                                            newValue = {
                                                "$uuid": "4e54a41e-5c7d-4926-ad57-6a5f882c24c4",
                                                "$key": "4e54a41e-5c7d-4926-ad57-6a5f882c24c4",
                                                "code": "AD",
                                                "description": "Andorra"
                                            };
                                    }
                                    data[$bind].splice(data[$bind].length > 1 ? 1 : 0, 0, newValue);
                                }
                                if (value.$reorderItem) {
                                    data[$bind].reverse();
                                }
                                fieldServerData.setDataValue(fieldServerData.getDataValue() + "\r\n" + JSON.stringify(data));
                                self.applyChange(data);
                            }
                        });
                    }
                });
            }
        }
        else {
            if (binding.indexOf("array") == 0) {
                var data = {
                    $etag: this.dataset.$etag || 1,
                    $url: this.getArticle().$prototype.$representationUrl,
                    binding: value
                };
                self.boundFields.$fieldServerData[0].setDataValue(JSON.stringify(data));
                return null;
            }
        }
        return FieldsTest.prototype.onNotifyRecordChange.call(this, value, binding);
    },
    _fillDiagnoses: function(value, binding){
        var metaData = FieldsTest.prototype._fillDiagnoses.call(this, value, binding);
        if (binding.indexOf("array") == 0) {
            var field = this.boundFields[binding][0];
            metaData.$items = [];
            var $diagnoseItem;
            if (value == "clear") {
                $diagnoseItem = null;
            }
            else {
                $diagnoseItem = [{
                    severity: "error",
                    message: "Error test for value"
                }];
                if (value == "full") {
                    $diagnoseItem.push({
                        severity: "warning",
                        message: "Warning test for value"
                    });
                    $diagnoseItem.push({
                        severity: "info",
                        message: "Info test for value"
                    });
                }
            }
            field.builder.getListDataSet().forEach(function(dataValue){
                metaData.$items.push({
                    $diagnoses: $diagnoseItem
                });
            });
        }
		return metaData;
    },
    loadBox: function(){
        var $prototype = {
            "$properties": {
                arrayString: {
                    "$type": "application/x-collection",
                    "$maxItems": 6,
                    "$minItems": 1,
                    "$title": "Array string",
                    "$item": {
                        "$type": "application/x-string",
                        "$constraints": {
                            "$maxLength": 20
                        }
                    }
                },
                arrayEmail: {
                    "$type": "application/x-collection",
                    "$maxItems": 4,
                    "$minItems": 1,
                    "$title": "Array Email",
                    "$item": {
                        "$type": "application/x-string",
                        "$format": "$email",
                        "$constraints": {
                            "$maxLength": 20
                        }
                    }
                },
                arrayChoice: {
                    "$type": "application/x-collection",
                    "$maxItems": 3,
                    "$minItems": 1,
                    "$title": "Array Choice",
                    "$item": {
                        "$type": "application/x-choice",
                        "$value": {
                            "$type": "application/x-integer",
                            "$constraints": {
                                "$enum": [{
                                    "$value": 1,
                                    "$title": "Value 1"
                                }, {
                                    "$value": 2,
                                    "$title": "Value 2"
                                }, {
                                    "$value": 3,
                                    "$title": "Value 3"
                                }, {
                                    "$value": 4,
                                    "$title": "Value 4"
                                }, {
                                    "$value": 5,
                                    "$title": "Value 5"
                                }]
                            }
                        }
                    }
                },
                arrayDate: {
                    "$type": "application/x-collection",
                    "$maxItems": 3,
                    "$minItems": 1,
                    "$title": "Array Date",
                    "$item": {
                        "$type": "application/x-date",
                        "$constraints": {
                            "$maxLength": 20
                        }
                    }
                },
                arrayBoolean: {
                    "$type": "application/x-collection",
                    "$maxItems": 3,
                    "$minItems": 1,
                    "$title": "Array Boolean",
                    "$item": {
                        "$type": "application/x-boolean",
                        "$constraints": {
                            "$maxLength": 20
                        }
                    }
                },
                arrayDecimal: {
                    "$type": "application/x-collection",
                    "$maxItems": 3,
                    "$minItems": 1,
                    "$title": "Array Decimal",
                    "$item": {
                        "$type": "application/x-decimal",
                        "$constraints": {
                            "$maxLength": 20
                        }
                    }
                },
                arrayReference: {
                    "$newControl": true,
                    "$type": "application/x-collection",
                    "$maxItems": 3,
                    "$minItems": 1,
                    "$title": "Array Reference",
                    "$item": {
                        $representation: "employee",
                        $title: "application/x-reference",
                        $isMandatory: false,
                        $capabilities: "sort,filter",
                        $type: "application/x-reference",
                        $value: "{code}",
                        $key: "{$uuid}",
                        "$properties": {
                            "code": {
                                "$type": "application/x-string"
                            },
                            "description": {
                                "$type": "application/x-string"
                            }
                        },
                        $links: {
                            "$details": {
                                "$type": "application/json;vnd.sage=syracuse",
                                "$url": "/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
                            },
                            "$lookup": {
                                "$type": "application/json;vnd.sage=syracuse",
                                "$url": "/sdata/sprint1/settings/sprint1/countries?representation=country.$lookup&role={$role}&trackingId={$trackingId}&binding=country"
                            }
                        }
                    }
                },
            }
        };
        this._appendServerBar($prototype);
        FieldsTest.prototype.loadBox.call(this, $prototype, {
            arrayString: ["dazhnda  d,aùe,dùe  daz dazed az", "deuxizem eada fa  ada", "troisième"],
            arrayEmail: ["toto@sage.com", "titi@sage.com", "tutu@sage.com"],
            arrayChoice: [2, 1, 5],
            arrayDate: ["2011-06-01", "2012-02-08", "2010-06-05"],
            arrayBoolean: [true, false, true],
            arrayDecimal: [2987111111111.4548, 248.8, 7635526.2],
            arrayReference: [{
                "$uuid": "4e54a41e-5c7d-4926-ad57-6a5f882c24c4",
                "$key": "4e54a41e-5c7d-4926-ad57-6a5f882c24c4",
                "code": "AD",
                "description": "Andorra"
            }],
            "$field$capabilityInsert": true,
            "$field$capabilityDelete": true,
            "$field$capabilityAppend": true,
            "$field$capabilityReorder": true
        
        });
    },
    appendSettingsBar: function(){
        return {
            $layoutType: "tabs",
            $items: [this._appendServerBar(), FieldsTest.prototype.appendSettingsBar.call(this)]
        };
    },
    _makeBlock: function($isEditMode){
        return [this._makeEditReadBlock("arrayString", [{
            $bind: "arrayString"
        }]), this._makeEditReadBlock("arrayEmail", [{
            $bind: "arrayEmail"
        }]), this._makeEditReadBlock("arrayChoice", [{
            $bind: "arrayChoice"
        }]), this._makeEditReadBlock("arrayDate", [{
            $bind: "arrayDate"
        }]), this._makeEditReadBlock("arrayBoolean", [{
            $bind: "arrayBoolean"
        }]), this._makeEditReadBlock("arrayDecimal", [{
            $bind: "arrayDecimal"
        }]), this._makeEditReadBlock("arrayReference", [{
            $bind: "arrayReference"
        }])];
    }
});
