// syracuse-ui/lib/common/tests/messageBoxTest.js
"use strict";
var helpers = require("syracuse-core/lib/helpers")
var Article = require("syracuse-ui/lib/common/article/article").Article;

function MessageBoxTest(){
}

exports.MessageBoxTest = helpers.defineClass(MessageBoxTest, Article, {
    loadBox: function(){
        var self = this;
        self.messageTypes = {
            alert: {
                $title: "This is an alert",
                $message: "Hi, I am an alert message\n with carriage return"
            },
            fatal: {
                $autoClose: true,
                $title: "This is a fatal message",
                $message: "Hi, I am a fatal  message",
                $type: "fatal"
            },
            warning: {
                $title: "This is a warning",
                $message: "Hi, I am a warning message",
                $type: "warning"
            },
            info: {
                $autoClose: 1000,
                $title: "This is an information",
                $message: "Hi, I am an information message",
                $type: "info"
            },
            question: {
                $title: "This is a question",
                $message: "Hi, I am a question",
                $type: "question",
                $buttons: "yesnocancel",
                callback: function(response){
                    alert(response.$id);
                }
            },
            prompt: {
                $title: "This is a prompt",
                $message: "Hi, I am a prompt",
                $type: "prompt",
                callback: function(response){
                    alert(response.prompt);
                    return true;
                }
            }
        };
        self.$prototype = {
            $links: {}
        };
        Object.keys(self.messageTypes).forEach(function($type){
            self.$prototype.$links[$type] = {
                $title: "Open popup"
            }
        });
        self.$item = {
            $skin: "s-portlet",
            $layout: {
                $items: [{
                    $layoutType: "columns",
                    $layoutSubType: "33,33,33",
                    $items: [{
                        $layoutType: "stack",
                        $items: self._makeSection(["alert", "fatal"])
                    }, {
                        $layoutType: "stack",
                        $items: self._makeSection(["info", "warning"])
                    }, {
                        $layoutType: "stack",
                        $items: self._makeSection(["question"])
                    }]
                }, {
                    $layoutType: "stack",
                    $items: self._makeSection(["prompt"])
                }]
            }
        }
        Article.prototype.loadBox.call(self);
        Object.keys(self.messageTypes).forEach(function($type){
            var section = self.idMap["messageBoxSection-" + $type];
            if ($type == "info") {
                section.$$title.text(section.$$title.text() + " ($autoClose = 1000)");
            }
            if ($type == "fatal") {
                section.$$title.text(section.$$title.text() + " ($autoClose = true 4s)");
            }
            var $msg = self.messageTypes[$type];
            $msg.$$container = section.$$body;
            document.site.showMessage($msg);
            delete $msg.$$container;
        });
    },
    _makeSection: function($types){
        return $types.map(function($type){
            return {
                $category: "section",
                $id: "messageBoxSection-" + $type,
                $title: "$type = '" + $type + "'",
                $layout: {
                    $items: [{
                        $category: "link",
                        $bind: $type
                    }]
                }
            }
        });
    },
    onMenuClick: function(menu){
        document.site.showMessage(this.messageTypes[menu.$bind]);
    }
});
