// syracuse-ui/lib/common/tests/referenceFieldsTest.js
"use strict";
var helpers = require("syracuse-core/lib/helpers")
var FieldsTest = require("./fieldsTest").FieldsTest;

function ReferenceFieldsTest(){
}

exports.ReferenceFieldsTest = helpers.defineClass(ReferenceFieldsTest, FieldsTest, {
    loadBox: function(){
        FieldsTest.prototype.loadBox.call(this, {
            "$properties": {
                "application/x-reference": {
                    $pluralType: "employees",
                    $baseUrl: "http://localhost:8126/sdata/sprint1/settings/sprint1",
                    $representation: "employee",
                    $title: "application/x-reference",
                    $isMandatory: false,
                    $capabilities: "sort,filter",
                    $type: "application/x-reference",
                    $url: "{$baseUrl}/{$pluralType}('{$key}')?representation={$representation}.$thumb",
                    $value: "{code}",
                    $key: "{$uuid}",
                    "$properties": {
                        "code": {
                            "$type": "application/x-string"
                        },
                        "description": {
                            "$type": "application/x-string"
                        }
                    },
                    $links: {
                        "$details": {
                            "$type": "application/json;vnd.sage=syracuse",
                            "$url": "/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
                        },
                        "$lookup": {
                            "$type": "application/json;vnd.sage=syracuse",
                            "$url": "/sdata/sprint1/settings/sprint1/countries?representation=country.$lookup&role={$role}&trackingId={$trackingId}&binding=country"
                        }
                    }
                },
                "application/x-array": {
                    "$title": "application/x-array application/x-reference",
                    "$capabilities": "sort,filter",
                    "$type": "application/x-array",
                    "$item": {
                        "$type": "application/x-reference",
                        "$url": "{$baseUrl}/{$pluralType}('{$key}')?representation={$representation}.$thumb",
                        "$value": "{description}",
                        "$description": "Access administration, groups associates users with roles and endpoints",
                        "$key": "{$uuid}",
                        "$properties": {
                            "description": {
                                "$type": "application/x-string"
                            }
                        },
                        "$links": {
                            "$details": {
                                "$type": "application/json;vnd.sage=syracuse",
                                "$url": "{$baseUrl}/groups('{$key}')?representation=group.$details&role={$role}"
                            },
                            "$select": {
                                "$type": "application/json;vnd.sage=syracuse",
                                "$title": "Ajouter",
                                "$url": "{$baseUrl}/groups?representation=group.$select&role={$role}&trackingId={$trackingId}&binding=groups"
                            }
                        }
                    }
                }
            }
        }, {
            "application/x-reference": {
                "$uuid": "4e54a41e-5c7d-4926-ad57-6a5f882c24c4",
                "$key": "4e54a41e-5c7d-4926-ad57-6a5f882c24c4",
                "$etag": 1,
                "$properties": {},
                "code": "AD",
                "description": "Andorra"
            },
            "application/x-array": [{
                "$uuid": "c8d848cc-d9eb-417b-8293-c576f91923a0",
                "$key": "c8d848cc-d9eb-417b-8293-c576f91923a0",
                "$etag": 1,
                "$properties": {},
                "description": "Overseas accountants",
                "explorer": {
                    "$url": "{$baseUrl}/{$pluralType}('{$key}')/$graphs/explorer",
                    "$type": "graph"
                },
                "parent": null,
                "role": {
                    "$uuid": "22c33896-bb4f-402a-9f73-f3e9824e035c",
                    "$key": "22c33896-bb4f-402a-9f73-f3e9824e035c",
                    "$etag": 1,
                    "$properties": {},
                    "description": "Accountant"
                }
            }, {
                "$uuid": "c97da1b5-c5b3-40e9-a7b3-c590decfab7b",
                "$key": "c97da1b5-c5b3-40e9-a7b3-c590decfab7b",
                "$etag": 1,
                "$properties": {},
                "description": "French accountants",
                "explorer": {
                    "$url": "{$baseUrl}/{$pluralType}('{$key}')/$graphs/explorer",
                    "$type": "graph"
                },
                "parent": null,
                "role": {
                    "$uuid": "22c33896-bb4f-402a-9f73-f3e9824e035c",
                    "$key": "22c33896-bb4f-402a-9f73-f3e9824e035c",
                    "$etag": 1,
                    "$properties": {},
                    "description": "Accountant"
                }
            }, {
                "$uuid": "f8a33a22-3d4c-41ca-b792-2fa6fe18f131",
                "$key": "f8a33a22-3d4c-41ca-b792-2fa6fe18f131",
                "$etag": 1,
                "$properties": {},
                "description": "Endpoint administrators",
                "explorer": {
                    "$url": "{$baseUrl}/{$pluralType}('{$key}')/$graphs/explorer",
                    "$type": "graph"
                },
                "parent": null,
                "role": {
                    "$uuid": "ab22fd7a-98fc-4ce8-90e8-f46a6462368f",
                    "$key": "ab22fd7a-98fc-4ce8-90e8-f46a6462368f",
                    "$etag": 1,
                    "$properties": {},
                    "description": "Endpoint administrator",
                    "functionProfile": null
                }
            }, {
                "$uuid": "6ea0e99b-2d22-4a8a-9128-d54336ac0ec8",
                "$key": "6ea0e99b-2d22-4a8a-9128-d54336ac0ec8",
                "$etag": 1,
                "$properties": {},
                "description": "Super administrators",
                "explorer": {
                    "$url": "{$baseUrl}/{$pluralType}('{$key}')/$graphs/explorer",
                    "$type": "graph"
                },
                "parent": null,
                "role": {
                    "$uuid": "f56e5600-a09d-4d4c-83fe-161c9ff13db2",
                    "$key": "f56e5600-a09d-4d4c-83fe-161c9ff13db2",
                    "$etag": 1,
                    "$properties": {},
                    "description": "Super administrator",
                    "functionProfile": null
                }
            }, {
                "$uuid": "8413d24d-27a2-444b-b5f2-387534618714",
                "$key": "8413d24d-27a2-444b-b5f2-387534618714",
                "$etag": 1,
                "$properties": {},
                "description": "zzzz",
                "explorer": {
                    "$url": "{$baseUrl}/{$pluralType}('{$key}')/$graphs/explorer",
                    "$type": "graph"
                },
                "parent": null,
                "role": {
                    "$uuid": "5bdd2df0-17b9-4fea-b5bb-23ec3c0180de",
                    "$key": "5bdd2df0-17b9-4fea-b5bb-23ec3c0180de",
                    "$etag": 1,
                    "$properties": {},
                    "description": "Admin",
                    "functionProfile": "Adlministrateur"
                }
            }]
        });
    },
    _makeBlock: function($isEditMode){
        return [this._makeEditReadBlock("application/x-string", [{
            $bind: "application/x-reference"
        }, {
            $bind: "application/x-array"
        }])];
    }
});
