// syracuse-ui/lib/common/tests/resources.js
"use strict";

exports.getRecords = function(count){
    var records = [];
    for (var index = 0; index < count; index++) {
        records.push({
            "$uuid": "df3f7fa4-7b64-4daa-b1f0-2222" + index,
            "$etag": 1,
            "$": {},
            "$index": index,
            "application/x-string": "record " + index,
            "application/x-choice": index,
            "application/x-date": "2011-0" + Math.max(Math.min(1, index), 9) + "-07",
            "application/x-boolean": true,
            "application/x-decimal": 45454.0 * index
        });
    }
    return records;
};


exports.getActionLinks = function(value){
    return {
        $empty: {
            $links: null
        },
        $lazyload: {
            $links: {
                "$lazyload": {}
            }
        },
        $links: {
            $links: {
                "uioverview": {
                    "$title": "UI Overview",
                    "$url": "?representation=ui.$dashboard"
                },
                "uilist": {
                    "$title": "UI List Overview",
                    "$url": "?representation=uilist.$dashboard"
                },
                "transitionTest": {
                    "$title": "Json Page test",
                    "$url": "?representation=uipagetest.$dashboard"
                },
                "$create": {
                    "$type": "application/json;vnd.sage=syracuse",
                    "$title": "Créer",
                    "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries/$template/$workingCopies?representation=country.$edit&role={$role}",
                    "$method": "POST"
                },
                "$print": {
                    "$type": "application/pdf",
                    "$title": "Imprimer",
                    "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries?representation=country.$query&role={$role}"
                },
                "$details": {
                    "$type": "application/json;vnd.sage=syracuse",
                    "$title": "Detail",
                    "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$details&role={$role}"
                },
                "$edit": {
                    "$type": "application/json;vnd.sage=syracuse",
                    "$title": "Edit",
                    "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')/$workingCopies?representation=country.$edit&role={$role}",
                    "$method": "POST"
                },
                "$delete": {
                    "$title": "Delete",
                    "$confirm": "The record  will be deleted. Confirm ?",
                    "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')?representation=country.$query&role={$role}",
                    "$type": "application/json;vnd.sage=syracuse",
                    "$method": "DELETE"
                },
                "$getSurface": {
                    "$type": "application/json;vnd.sage=syracuse",
                    "$title": "Get surface",
                    "$method": "GET",
                    "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')/$service/$getSurface"
                },
                "$getPopulation": {
                    "$type": "application/json;vnd.sage=syracuse",
                    "$title": "Get population",
                    "$method": "GET",
                    "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')/$service/$getPopulation"
                },
                "$getPoliticalRegime": {
                    "$type": "application/json;vnd.sage=syracuse",
                    "$title": "Political regime",
                    "$method": "GET",
                    "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')/$service/$getPoliticalRegime"
                },
                "$getNumber4": {
                    "$type": "application/json;vnd.sage=syracuse",
                    "$title": "Et de 4",
                    "$method": "GET",
                    "$url": "http://localhost:8126/sdata/sprint1/settings/sprint1/countries('{$uuid}')/$service/$getNumber4"
                }
            }
        },
        $actions: {
            $actions: {
                $save: {
                    $title: "Save",
                    $isDisabled: true
                }
            }
        },
        $linksParameters: {
            $links: {
                $linksExpressionParameters: {
                    $title: "Expression parameters",
                    $description: "param2 from xstring,param2 from json",
                    $url: "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&param1={param1}&param2={param2}",
                    $parameters: {
                        param1: "application/x-string",
                        param2: "value in json"
                    }
                },
                $linksFormParameters: {
                    $title: "Form parameters",
                    $description: "param1 and param2 from default, param3,param4 and param5 from form",
                    $url: "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query&par1={param1}&par2={param2}&par3={param3}&par4={param4}&par4={param5}",
                    $parameters: {
                        "$": {
                            param3: {
                                $title: "mon param3",
                                "$type": "application/x-string"
                            },
                            "param4": {
                                "$title": "mon param4",
                                "$type": "application/x-string"
                            },
                            "param5": {
                                "$title": "mon param4",
                                "$type": "application/x-string",
                                "$isEditMode": false
                            }
                        },
                        "param1": "application/x-string",
                        "param2": "value"
                    }
                }
            }
        
        },
        $actionsParameters: {
            $actions: {
                "actionEmployeesProperties": {
                    "$title": "Action Employees Properties",
                    "$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query",
                    "$parameters": {
                        "param1": "{lastName}",
                        "param2": "value"
                    }
                },
                "actionEmployeesPropertiesForm": {
                    "$title": "Action Employees Properties Form",
                    "$url": "/sdata/sprint1/settings/sprint1/employees?representation=employee.$query",
                    "$parameters": {
                        "$": {
                            "param3": {
                                "$title": "{@S-AOC0-125}",
                                "$type": "application/x-string"
                            },
                            "param4": {
                                "$title": "{@S-AOC0-8}",
                                "$type": "application/x-string"
                            },
                            "param5": {
                                "$title": "{@S-AOC1-1390}",
                                "$type": "application/x-string",
                                "$isEditMode": false
                            }
                        },
                        "param1": "{lastName}",
                        "param2": "value"
                    }
                }
            }
        }
    
    }[value];
};
