// syracuse-ui/lib/common/tests/simpleFieldsTest.js
"use strict";
var helpers = require("syracuse-core/lib/helpers")
var FieldsTest = require("./fieldsTest").FieldsTest;

function SimpleFieldsTest(){
}

exports.SimpleFieldsTest = helpers.defineClass(SimpleFieldsTest, FieldsTest, {
    loadBox: function(){
        FieldsTest.prototype.loadBox.call(this, {
            "$properties": {
                "application/x-string": {
                    $title: "x-string",
                    $type: "application/x-string",
                    
                        $maxLength: 30
                    ,
                    $capabilities: "sort,filter,alphaTab",
                    $links: {
                        $tunnel: {},
                        $lookup: {}
                    }
                },
                "application/x-choice": {
                    $title: "application/x-choice",
                    $type: "application/x-choice",
                    $value: {
                        $type: "application/x-integer",
                            $enum: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10].map(function(value){
                                return {
                                    "$value": value,
                                    $title: "Value " + value
                                };
                            })
                    },
                    $links: {
                        $tunnel: {},
                        $lookup: {}                        
                    }
                },
                "application/x-date": {
                    $title: "x-date",
                    $type: "application/x-date",
                    $links: {
                        $tunnel: {},
                        $lookup: {}
                    }
                },
                "application/x-time": {
                    $title: "x-time",
                    $type: "application/x-time"
                },
                "application/x-datetime": {
                    $title: "x-datetime",
                    $type: "application/x-datetime"
                },
                "application/x-boolean": {
                    $title: "application/x-boolean",
                    $type: "application/x-boolean"
                },
                "$email": {
                    $title: "$format: $email",
                    $type: "application/x-string",
                    $format: "$email"
                
                },
                "$phone": {
                    $title: "$format: $phone",
                    $type: "application/x-string",
                    $format: "$phone"
                },
                "application/x-integer": {
                    $title: "application/x-integer",
                    $type: "application/x-integer",
                    
                        "$isNullable": true
                    ,
                    "$capabilities": "sort,filter",
                    $links: {
                        $tunnel: {}
                    }
                },
                "application/x-decimal": {
                    $title: "application/x-decimal",
                    $type: "application/x-decimal",
                    
                        "$isNullable": true
                    },
                "image": {
                    $title: "image",
                    $type: "image"
                }
            }
        }, {
            "application/x-string": "ceci est du texte",
            "$phone": "01.41.66.21.21",
            "$email": "toto@sage.com",
            "application/x-integer": 200,
            "application/x-decimal": 2987111111111.4548,
            "application/x-boolean": true,
            "application/x-date": "2011-06-07",
            "application/x-time": "12:00:00",
            "application/x-datetime": "2012-03-16T12:40:50.957Z",
            "application/x-choice": 2
        });
    },
    _makeBlock: function($isEditMode){
        return [this._makeEditReadBlock("application/x-string", [{
            $bind: "application/x-string"
        }, {
            $bind: "$phone"
        }, {
            $bind: "$email"
        }]), this._makeEditReadBlock("application/x-boolean", [{
            $bind: "application/x-date"
        }, {
            $bind: "application/x-time"
        }, {
            $bind: "application/x-datetime"
        }]), this._makeEditReadBlock("Numeric", [{
            $bind: "application/x-integer"
        }, {
            $bind: "application/x-decimal"
        }]), this._makeEditReadBlock("application/x-boolean", [{
            $bind: "application/x-boolean"
        }]), this._makeEditReadBlock("application/x-choice", [{
            $title: "default >4 combo",
            $bind: "application/x-choice"
        }, {
            $title: "$format:$combo",
            $bind: "application/x-choice",
            $format: "$combo"
        }, {
            $title: "$format:$radios",
            $bind: "application/x-choice",
            $format: "$radios"
        }]), this._makeEditReadBlock("image", [{
            $title: "image",
            $bind: "image"
        }])];
    }
});
