// syracuse-ui/lib/controller/controller.js
"use strict";
var _helpers = require('syracuse-core/lib/helpers');
var _soapPage = require('syracuse-ui/lib/page/soap/soapPage');

var _mainPageFetching;

function _openTracker(menu, menuArticle, $url, initData) {
	if (menu.$invocationMode == "async" && menuArticle.dataset && menuArticle.dataset.$trackngId) {
		initData = initData || {};
		initData.$location = $url;
		initData.title = (menu.getTitle && menu.getTitle()) || menu.$title;
		initData.uuid = initData.uuid || menuArticle.dataset.$trackngId;
		var page = menuArticle.page;
		if (page.openerUrlSegments.$facet != "$edit") {
			(initData.$links = initData.$links || {}).$opener = {
				$title: page.getTitle(),
				$url: page.openerUrlSegments.$url
			};
		}
		syra_trackers.addTracker(initData);
		return true;
	}
}



function _attachUrlWithRepresentation(segment, $url) {
	if (segment.params) {
		if (segment.params.profile || segment.params.role) {
			var urlSegments = syra_url.parse($url);
			if (segment.params.role) {
				urlSegments.params.role = segment.params.role;
			}
			if (segment.params.profile) {
				urlSegments.params.profile = segment.params.profile;
			}
			$url = syra_url.build(urlSegments);
		}
	}
	return $url;
}

function _fetchPage($representation, context, userPreferenceLoading) {
	$representation.$prototype = $representation.$prototype || {};
	if (typeof($representation.$prototype) == "string") {
		context.sendRqtOpts.$url = _attachUrlWithRepresentation(context.segments, $representation.$prototype);
		context.sendRqtOpts.success = function($prototype) {
			$representation.$prototype = $prototype;
			_fetchPage($representation, context);
		};
		syra_http.get(context.sendRqtOpts);
		return;
	}
	if ($representation.$prototype.$article && typeof($representation.$prototype.$article) == "string") {
		context.sendRqtOpts.$url = _attachUrlWithRepresentation(context.segments, $representation.$prototype.$article);
		context.sendRqtOpts.success = function($article) {
			$representation.$prototype.$article = $article;
			_fetchPage($representation, context);
		};
		syra_http.get(context.sendRqtOpts);
		return;
	}
	if (!userPreferenceLoading) {
		var $userPreferencesLink = $representation.$links && $representation.$links.$userPreferences;
		if ($userPreferencesLink) {
			if (!(context.segments && context.segments.representationRoot == "userProfile")) {
				context.sendRqtOpts.$url = $userPreferencesLink.$url;
				context.sendRqtOpts.success = function($newPreferences) {
					$representation.$newPreferences = $newPreferences;
					_fetchPage($representation, context, true);
				};
				syra_http.get(context.sendRqtOpts);
				return;
			}
		}
	}
	if ($representation.$article && typeof($representation.$article) == "string") {
		context.sendRqtOpts.$url = _attachUrlWithRepresentation(context.segments, $representation.$article);
		context.sendRqtOpts.success = function($article) {
			$representation.$article = $article;
			_fetchPage($representation, context);
		};
		syra_http.get(context.sendRqtOpts);
		return;
	}

	syra_site.userProfile && syra_site.userProfile.setSecurityToRepresentation($representation);
	$representation.$prototype.$representationUrl = context.$representationUrl;
	if (context.onBeforeOpen) {
		context.onBeforeOpen($representation);
	}
	context.success({
		openerUrlSegments: context.segments,
		$representation: $representation
	});
	syra_site.checkMainPageBookmarkStatus();
}

function _openDevPage(segments) {
	syra_site.loadStyleSheet("uitest.css");
	require.async("syracuse-ui/lib/tests/pages/devPage", function(err, module) {
		try {
			var $itemPage = {
				layoutSlot: document.createElement("div"),
				$category: "page",
				openerUrlSegments: segments,
				$pageCategoryClass: module.DevPage,
				$representation: {
					$prototype: {
						$devName: segments.representationRoot,
						$properties: {},
						$links: {},
						$actions: {}
					},
					$article: {}
				}
			};
			$itemPage.openerUrlSegments.fullUrl = document.location.href;
			syra_site.onMainPageChange($itemPage);
		} catch (error) {
			console.error(error.message + "\n" + error.stack);
		}
	});
}

exports.changeMainPage = function($url, $itemPage, onHistoryChange) {
	var designer = syra_site.mainPage && syra_site.mainPage.designer;
	if (designer && designer.onBeforeMainPageChange) {
		if (!designer.onBeforeMainPageChange(function() {
			exports.changeMainPage($url, $itemPage);
		})) {
			return false;
		}
	}
	if (_mainPageFetching) {
		var ajaxQuery = _mainPageFetching.sendRqtOpts && _mainPageFetching.sendRqtOpts.ajaxQuery;
		ajaxQuery && ajaxQuery.abort();
		_mainPageFetching = null;
	}
	var workingCopyPreload = syra_form.preload;
	syra_form.preload = null; //!important release buffer
	var openerUrlSegments = syra_url.parse($url);

	if (openerUrlSegments.isFusion) {
		syra_fusion.openMainPage(openerUrlSegments);
	} else {
		if (syra_fusion && syra_fusion.activatedBook) {
			syra_fusion.isClosingAll = true;
			syra_fusion.activatedBook.onCloseAllSheets(function(allSheetsClosed) {
				delete syra_fusion.isClosingAll;
				if (!allSheetsClosed) {
					onHistoryChange && syra_url.history.cancelChange();
				} else {
					exports.changeMainPage($url, $itemPage);
				}
			});
			return false;
		}
		if (openerUrlSegments.$url.indexOf("soap-") >= 0) {
			_soapPage.open(openerUrlSegments);
			return;
		}
		if (openerUrlSegments.$facet == "$dev") {
			_openDevPage(openerUrlSegments);
			return;
		}

		if (syra_quality.openTestPage(openerUrlSegments)) {
			return;
		}

		if (syra_site.controller_onChangeMainPage && !syra_site.controller_onChangeMainPage(openerUrlSegments)) {
			return;
		}

		if ($itemPage) {
			$itemPage.initData = $itemPage.initData || (workingCopyPreload && workingCopyPreload.initData);
			$itemPage.openerUrlSegments = openerUrlSegments;
			$itemPage.openerUrlSegments.fullUrl = document.location.href;
			syra_site.onMainPageChange($itemPage);
		} else {
			exports.loadRepresentation(_mainPageFetching = {
				article: null,
				segments: openerUrlSegments,
				success: function($itemPage) {
					_mainPageFetching = null;
					$itemPage.initData = $itemPage.initData || (workingCopyPreload && workingCopyPreload.initData);
					$itemPage.openerUrlSegments = openerUrlSegments;
					if (workingCopyPreload) {
						$itemPage.openerUrlSegments.workingCopyPreloadUrl = workingCopyPreload.openUrl;
						// save working copy preload url (session storage)
						sessionStorage &&
							sessionStorage.setItem(openerUrlSegments.$url, JSON.stringify({
								wcpreloadUrl: workingCopyPreload.openUrl
							}));
					}
					$itemPage.openerUrlSegments.fullUrl = document.location.href;
					syra_site.onMainPageChange($itemPage);
				},
				error: function(error, httpQuery) {
					_mainPageFetching = null;
					if (this.isAborted) {
						return false;
					}
					syra_site.onErrorMainpage = true;
					return syra_diagnose.showPage("show" + error.status, error, httpQuery.extractDiagnoses(error));
				}
			});

		}
	}
};

//article, segments, success, onError, onBeforeOpen, sendRqtOpts
exports.loadRepresentation = function(context) {
	if (typeof(context.segments) == "string") {
		context.segments = syra_url.parse(context.article ? syra_expression.parse(context.article, context.segments) : context.segments);
	}
	context.$representationUrl = context.segments.$url;
	//xhr.withCredentials = true;
	context.sendRqtOpts = context.sendRqtOpts || {};
	context.sendRqtOpts.$url = syra_url.buildPageCollaborationUrl(context.article || syra_site, context.segments);
	context.sendRqtOpts.error = context.onError;
	context.sendRqtOpts.success = function($representation, response, $url) {
		$representation.$pageCollaborationUrl = $url;
		_fetchPage($representation, context);
	};
	syra_http.get(context.sendRqtOpts);
};

exports.postQuery = function(menu, record, article, callback, callbackError) {
	var menuArticle = article || menu.articleParent || syra_site;
	menu.$method = menu.$method || "POST";
	syra_url.formatMenuUrl(menuArticle, menu, record, function($url) {
		menu.$url = $url;
		var params = {};
		if (menu.$invocationMode === "async" && menu.$trackngId) {
			params.trackngId = menu.$trackngId;
			if (menu.$capabilities && menu.$capabilities.indexOf("abort") >= 0) {
				params.canAbort = true;
			}
		} else {
			params.trackingId = _helpers.uuid.generate();
		}
		syra_http.send({
			page: menuArticle,
			$url: $url,
			$acceptType: menu.$acceptType || menu.$type,
			$contentType: menu.$contentType,
			params: params,
			headers: menu.httpHeaders,
			method: menu.$method,
			data: menu.sendData,
			noDisplayErr: menu.noDisplayErr || false,
			success: function(data, response, requestUrl) {
				var $location = {
					$url: response.headers.location,
					type: ""
				};
				if ($location.$url) {
					$location.$type = response.headers["content-type"];
				} else {
					$location.$url = response.data.$url || menu.$url;
					$location.$type = response.data.$type || menu.$type;
				}
				if (menu.$invocationMode === "async") {
					_openTracker(menu, menuArticle, $location.$url, response.data);
				} else {
					callback($location, response.data, requestUrl);
				}
			},
			error: function(err) {
				var $diagnoses = err.data && err.data.length > 0 ? (err.data.indexOf("$diagnoses") >= 0 ? JSON.parse(err.data).$diagnoses : [{
					$message: err.data ? err.data : syra_local.http_empty_error
				}]) : [{
					$message: syra_local.default_err_msg
				}];
				if (callbackError) {
					callbackError($diagnoses, err);
				} else {
					syra_diagnose.showDiagnoses({
						$diagnoses: $diagnoses
					});
				}
			}
		});
	});
};

exports.executeMenu = function(menu, article, confirmed) {
	if (menu.$isAction || menu.$url || menu.$sourceUrl) {
		var menuArticle = article || menu.articleParent;
		var menuTitle = (menu.getTitle && menu.getTitle()) || menu.$title;
		if (menu.$url && menu.$url == "{$url}") { //use by search
			menu.$url = syra_expression.parse(menuArticle, menu.$url);
		}
		if (menu.$method && menu.$method.indexOf("{") >= 0) { //use by search
			menu.$method = syra_expression.parse(menuArticle, menu.$method);
		}
		if (menuArticle && menuArticle.dataset) {
			if (menu.$invocationMode === "async") {
				menu.$trackngId = menuArticle.dataset.$trackngId = _helpers.uuid.generate();
			} else {
				delete menu.$trackngId;
				delete menuArticle.dataset.$trackngId;
			}
		}
		// temporarily, until $confirm-message protocol extension
		if (menu.$confirm && menu.$officeAddinSetup) {
			if (_helpers.getCookie("SyracuseOfficeAddin") === "valid") {
				menu.$confirm = "";
			}

		}
		if (menu.$url && menu.$url.indexOf("SyracuseOfficeAddinsSetup.EXE") > 0) {
			_helpers.setCookie("SyracuseOfficeAddin", "valid", null, (new Date((new Date().getYear() + 1910), new Date().getMonth(), new Date().getDate())).toUTCString());
		}
		if (!confirmed && menu.$confirm) {
			syra_diagnose.box.show({
				$title: syra_expression.parse(menuArticle, menuTitle || syra_local.siteDefaultConfirmTitle),
				$message: syra_expression.parse(menuArticle, menu.$confirm),
				$type: "question",
				callback: function(response) {
					if (response && response.$clientId === "yes") {
						// temporarily, until $confirm-message protocol extension
						if (menu.$officeAddinSetup) {
							_helpers.setCookie("SyracuseOfficeAddin", "valid", null, (new Date((new Date().getYear() + 1910), new Date().getMonth(), new Date().getDate())).toUTCString());
						}
						// temporarily END
						exports.executeMenu(menu, article, true);
					}
				}
			});
			return;
		}
		if (menu.$isAction) {
			var target = menu.contextField || menuArticle;
			if (target || menuArticle) {
				var request = {};

				var r = request[menu.$sourceBind || (menu.$item && menu.$item.$bind)] = {
					$isRequested: true
				};
				if (menu.$invocationMode === "async") {
					r.$trackingId = _helpers.uuid.generate();
				}
				if (menu.$parameters) {
					syra_url.notifyParametersActionChange(menuArticle, menu, null, target, request);
				} else {
					menuArticle.page.notifyActionChange(menu, target, request);
				}
			}
		} else {
			if (menu.$method && (menu.$method != "GET")) {
				if (syra_form.isUrl(menu.$url, menu.$sourceUrl)) {
					syra_form.onBeforeUnloadPage(function() {
						syra_form.load({
							menu: menu,
							menuArticle: menuArticle,
							article: article
						});
					});
				} else {
					var page = menu.page;
					//var menuArticle = menu.articleParent;
					exports.postQuery(menu, null, null, function($location, data) {
						switch (page.$facet) {
							case "$query":
							case "$bulk":
								page.fetchLastRequest(function(lastData) {
									if (lastData && data.$diagnoses) {
										lastData.$diagnoses = lastData.$diagnoses || [];
										for (var ii = data.$diagnoses.length - 1; ii >= 0; ii--) {
											lastData.$diagnoses.unshift(data.$diagnoses[ii]);
										}
									}
									page.startChange();
									page.applyChange(lastData);
									page.endChange();
								});
								break;
							case "$details":
								page.applyChange(data);
								if (menu.$method == "DELETE") {
									page.freezeDeletedPage();
								}
								break;
							default:
								if (page) {
									page.startChange();
									page.onExecuteMenuResponse && page.onExecuteMenuResponse(menu.$sourceBind, menuArticle, data);
									page.endChange();
								}
								break;
						}
					});
				}
			} else {
				syra_form.onBeforeUnloadPage(function() {
					_openPage(menuArticle, menu);
				});
			}
		}
	}
};

function _openPage(menuArticle, menu, record) {
	var menuTitle = (menu.getTitle && menu.getTitle()) || menu.$title;
	switch (menu.$target) {
		case "worksheet":
			syra_site.excelDocument.loadPage({
				$url: menu.$url,
				$title: (menuArticle && menuArticle.dataset && menuArticle.dataset.title) || menuTitle
			});
			break;
		case "popup":
			syra_dlg.openPage(menu.boxParent, {
				article: menuArticle,
				$url: menu.$url,
				isModal: false
			});
			break;
		default:
			syra_url.formatMenuUrl(menuArticle, menu, record, function($url, isCanceled) {
				if (!isCanceled) {
					if (_openTracker(menu, menuArticle, $url)) {
						return;
					}
					var isRepresentationUrl = true;
					var segments = syra_url.parse($url, true);
					var target = menu.$target;
					if (target == "diagnoses") {
						syra_http.get({
							page: menuArticle,
							$url: $url,
							success: function(data, response, requestUrl) {
								var page = menuArticle && menuArticle.page;
								if (page) {
									page.startChange();
									page.applyChange(data, response, requestUrl);
									page.endChange();
								}
							}
						});
						return;
					}
					if (segments.isSyracuse) {
						if (segments.uri.search(/mobile\.html/i) > 0) {
							syra_url.history.load({
								$url: $url,
								target: "blank",
								isDefaultUrl: false
							});
							return;
						}
						if (segments.$facet == "$summary") {
							syra_dlg.openPage(menu.boxParent, {
								article: menuArticle,
								$url: $url,
								isModal: false
							});
							return;
						} else {
							if (menu.$type && menu.$type.indexOf("json") == -1) {
								target = "blank";
							} else {
								var htmIndex = segments.uri.search(/\.html/i);
								if (htmIndex > 0) {
									//test for excelOfficeConfig.html
									if ((segments.uri.length - 5) == htmIndex) {
										var curUrlSegments = syra_url.parse(document.location.href);
										if (curUrlSegments.uri.indexOf(segments.uri) < 0) {
											isRepresentationUrl = false;
											target = "blank";
										}
									}
								}
								if (isRepresentationUrl) {
									if (menu.boxParent) {
										if (menu.page && menu.page.dialogWrapper && menu.page.dialogWrapper.isModal) {
											target = "blank";
										}
									}
								}
							}
						}
					} else {
						target = "blank";
					}
					if (menu.page && menu.page.inlinePageHost) {
						if (!target || target == "self") {
							target = "blank";
						}
					}
					if (target != "blank" && target != "main" && menuArticle.page && menuArticle.page.$pageCategory == "vignette") {
						menuArticle.page.vignetteField.renderLayoutContent($url);
					} else {
						if (isRepresentationUrl && menu && (menu.isRepresentationUrl === false)) {
							isRepresentationUrl = false;
						}
						syra_url.history.load({
							replaceState: menu.replaceState,
							$url: $url,
							target: target,
							isDefaultUrl: isRepresentationUrl
						});
					}
				}
			});
			break;
	}
};
