// syracuse-ui/lib/controller/form.js
"use strict";
var _sendBag = require('./sendBag');

var _lastSavingDiagnoses;

//data.wcpreloadUrl = data.wcpreloadUrl || page.openerUrlSegments.workingCopyPreloadUrl;

var _storage = {
	hasData: function(page) {
		var data = sessionStorage && sessionStorage.getItem(page.openerUrlSegments.$url);
		if (data) {
			data = JSON.parse(data);
			if (data && data.dataset) {
				return true;
			}
		}
	},
	write: function(page) {
		// saving data for "404 working copy session expired" cases
		if (page.isWorkingCopy && sessionStorage && page.openerUrlSegments) {
			var key = page.openerUrlSegments.$url;
			var data = sessionStorage.getItem(key) !== "" ? JSON.parse(sessionStorage.getItem(key)) || {} : {};
			data.wcpreloadUrl = data.wcpreloadUrl || page.openerUrlSegments.workingCopyPreloadUrl;
			data.dataset = page.dataset;
			sessionStorage.setItem(key, JSON.stringify(data));
		}
	},
	read: function(dataObj, $url, reqOptions) {
		var data = sessionStorage && sessionStorage.getItem($url);
		if (data) {
			data = JSON.parse(data);
			var segs = syra_url.parse(reqOptions.$url = data.wcpreloadUrl);
			reqOptions.$url = syra_url.build(segs);
			reqOptions.data = dataObj || data.dataset;
			reqOptions.$etag = reqOptions.data ? reqOptions.data.$etag : reqOptions.$etag;
			return true;
		}
	},
	clear: function(page) {
		sessionStorage && sessionStorage.removeItem(page.openerUrlSegments.$url);
	}
};

exports.onWorkingCopyNotFound = function(httpQuery, error, dataObj) {
	var segments = syra_site && syra_site.mainPage && syra_site.mainPage.openerUrlSegments;
	var options = {};
	if (segments) {
		if (segments.workingCopyPreloadUrl) {
			options.$url = segments.workingCopyPreloadUrl;
			options.data = dataObj || syra_site.mainPage.dataset;
			options.$etag = syra_site.mainPage.dataset.$etag;
		} else {
			if (!_storage.read(dataObj, segments.$url, options)) {
				httpQuery.showError(error);
			}
		}
		options.page = syra_site.mainPage;
		options.success = httpQuery.options.success;
		options.error = httpQuery.options.error;
		syra_ajax.post(options);
	} else {
		httpQuery.showError(error);
	}
};

function _validateField(field, newValue, errors) {
	if (field.$item.$isFilterMode) {
		return true;
	}
	if (field.isArrayField) {
		var ok = true;
		for (var ii = 0, jj = field.records.length; ii < jj; ii++) {
			ok = exports.validate(field.records[ii]);
		}
		return ok;
	}
	if (field.isChildField) {
		return exports.validate(field.record);
	}

	var value = newValue;
	if (value === undefined) {
		value = field.currentValue;
	}
	if (value != null && value.toString) {
		value = value.toString();
	}
	errors = errors || [];
	if (!field.page.isFusionPage && field.$isMandatory) {
		var checked = value;
		if (field.isReferenceField) {
			checked = field._hasUiid ? field.currentValue && field.currentValue.$uuid : field.getInputValue();
		}
		if (checked === '' || checked === null || (!field.isPasswordField && checked === undefined)) {
			errors.push(syra_local.fieldIsMandatory);
		}
	}

	if (field.$field.$pattern && value) {
		var match = new RegExp(field.$field.$pattern, "i").test(value);
		if (!match) { //(match && match.index == 0 && match[0].length == value.length)) {
			errors.push(field.$field.$patternMessage || syra_local.fieldInvalidValue + " : " + value);
		}
	}

	field.validate && field.validate(errors, value);

	//alertPanel
	if (errors.length > 0 || field.isInvalid) {
		syra_fields.invalidate(field, errors);
		var saveLinks = field.articleParent.menuItems.$save;
		if (saveLinks && saveLinks.length > 0) {
			var hasErrors = field.page.alertPanel && field.page.alertPanel.hasErrors();
			if (!saveLinks[0].$isDisabled) {
				hasErrors && saveLinks[0].disable(true);
			} else {
				!hasErrors && saveLinks[0].disable(true);
			}
		}
		var sendBag = _sendBag.get(field.page);
		if (sendBag) {
			newValue = field.parentVariantField ? field.parentVariantField.saveVariantValue(newValue) : newValue;
			sendBag.save(field.page, field.articleParent, field.$item.$bind, newValue);
		}
	}
	return errors.length == 0;
};

exports.validate = function(article) {
	var isValidated = true;
	try {
		var page = article.page;
		if (page == article) {
			page.isPageValidating = page.isServerNotifyDisabled = true;
		}

		var $binds = Object.keys(article.boundFields);
		for (var ii = 0, jj = $binds.length; ii < jj; ii++) {
			var fields = article.boundFields[$binds[ii]];
			for (var kk = 0, mm = fields.length; kk < mm; kk++) {
				var field = fields[kk];
				if (!field.$isHidden && field.$isEditMode) {
					field = field.variantItem || field;
					if (!field.unknowMode) {
						var newValue;
						try {
							newValue = field.getValue();
							if (!_validateField(field, newValue)) {
								isValidated = false;
							} else {
								if (newValue != undefined) {
									if (page.isDialogParameter || (page.isPageValidating && newValue != field.currentValue)) {
										exports.update(field, newValue);
									}
								}
							}
						} catch (error) {
							isValidated = false;
							syra_fields.invalidate(field, error.message);
						}
					}
				}
			}
		}
	} finally {
		if (page == article) {
			page.isPageValidating = page.isServerNotifyDisabled = false;
		}
	}
	return isValidated;
};

exports.onInputChange = function(field, event) {
	field = field.variantItem || field;
	if (!field.unknowMode) {
		field.setDirty(true);
		try {
			var doValidate = true;
			if (field.isReferenceField) {
				doValidate = false;
				var value = field.input.value || "";
				if (value == "") {
					if (field.refDescriptionItem) {
						if (!field._titleField) {
							field.refDescriptionItem.textContent = "";
							syra_dom.hide(field.refDescriptionItem, true);
						}
					} else {
						field.input.title = "";
					}
					exports.update(field, null);
				} else {
					if ((field.currentValue && field.currentValue.$value != value) || !field.currentValue) {
						var postValue = {};
						postValue[field.$reference.$value.$itemProp] = value;
						exports.update(field, postValue);
					}
				}
			} else {
				switch (field.$field.$type) {
					case "application/x-integer":
					case "application/x-decimal":
					case "application/x-real":
					case "application/x-quantity":
						doValidate = false;
						var val = field.input.value,
							valFormatted = val;
						field.valHasChanged = true;
						if (val != '') {
							field.input.value = valFormatted = field._formatInput(field, val);
						}
						if (_validateField(field, valFormatted, [])) {
							doValidate = true;
						}
						break;
					case "application/x-choice":
						if (field.isRadioChoiceField) {
							event && event.target && field.selectRadioBtn(event.target.syraEnumIndex, true);
						} else {
							field.fillSelectedEnum(field.findEnum(field.getValue()));
						}
						break;
					case "application/x-password":
						if (field.confirmInput) {
							doValidate = false;
							if (event.target == field.confirmInput) {
								if (_validateField(field, field.confirmInput.value || "")) {
									field.currentValue = "";
									doValidate = true;
								}
							}
						}
						break;
				}
			}
			if (doValidate) {
				var newValue = field.getValue(event);
				if (newValue != field.currentValue) {
					// apply format  - SAM 106180
					if (field.$field.$x3Format) {
						newValue = syra_fields.format(newValue, field.$field.$x3Format, field.$field.$type, field.page.$prototype.$x3FormatExt);
					}
					exports.update(field, newValue);
				}
			}
		} catch (error) {
			syra_fields.invalidate(field, error.message);
			field.currentValue = newValue;
		}
	}
};

function _updateWorkingCopy(page, sendBag) {
	if (!page.disposed) {
		if (page.updateWorkingCopy ? page.updateWorkingCopy(sendBag) : true) {
			if (!page.isServerNotifyDisabled && page.$prototype.$representationUrl) {
				sendBag.$url = page.getDataUrl();
				sendBag.$etag = page.dataset.$etag;
				syra_ajax.put({
					page: page,
					url: page.getDataUrl(),
					data: sendBag,
					$etag: sendBag.$etag,
					success: function(data, response, requestUrl) {
						if (!page.disposed) {
							if (sendBag) {
								delete sendBag.$actions;
							}
							page.applyChange(data, response, requestUrl);
						}
					}
				});
			}
		}
	}
};

exports.update = function(field, newValue) {
	field = field.variantItem || field;
	if ((!field.unknowMode && !field.page.isFusionPage) || field.useDefaultUpdate) {
		field.currentValue = newValue;
		if (_validateField(field, newValue)) {
			if (field.notifyFieldChange ? field.notifyFieldChange(newValue) : true) {
				syra_dataset.setFieldValue(field, field.currentValue);
				exports.updateDelta(field, field.currentValue);
			}
		}
	}
};

exports.updateDelta = function(field, value) {
	var page = field.page;
	if (page) {
		if (field.articleParent.onFormUpdateDelta ? field.articleParent.onFormUpdateDelta(field, value) : true) {
			if (field.page.onFormUpdateDelta ? field.page.onFormUpdateDelta(field, value) : true) {
				var sendBag = _sendBag.get(page);
				sendBag.save(page, field.articleParent, field.$item.$bind, field.parentVariantField ? field.parentVariantField.saveVariantValue(value) : value);
				_updateWorkingCopy(page, sendBag);
				_storage.write(page);
			}
		}
		if (field.isBooleanField && field.input) {
			syra_fields.checkbox.setIcon(field, (value || field.statusValues.off) == field.statusValues.on);
		}
		if (field.isRadioChoiceField) {
			if (field._choices) {
				var selected = field.findEnum(value);
				selected && field.selectRadioBtn(selected.$index, true);
			}
		}
	}
};

exports.getSendBag = function(article, sender) {
	return _sendBag.get(article, sender);
};

exports.onBeforeUnloadPage = function(callback, activePage) {
	var wcPage = syra_site.activeWorkingCopy;
	if (wcPage && wcPage.$menus && wcPage.$menus.$save && _storage.hasData(wcPage)) {
		if (!activePage || (!activePage.isPageDesigner && !activePage.isAlertDialog && wcPage == activePage)) {
			syra_alert.ask({
				$title: wcPage.getTitle() || syra_local.wc_warn_unload_title,
				$message: syra_local.wc_warn_unload_message,
				$buttonsTitle: {
					no: syra_local.wc_warn_unload_no_button,
					yes: syra_local.wc_warn_unload_yes_button
				},
				callback: function(response) {
					if (response.$clientId == "yes") {
						wcPage.inlinePageHost && wcPage.inlinePageHost.dispose();
						callback();
					}
				}
			});
			return;
		}
	}
	callback();
};

exports.isFormUrl = function($url, $sourceUrl) {
	var ok = $url && $url.indexOf("$workingCopies") >= 0;
	if (!ok && $sourceUrl) {
		ok = $sourceUrl.indexOf("$workingCopies") >= 0;
	}
	return ok;
};


exports.load = function(options) {
	syra_router.postQuery(options.menu, options.record || null, options.menuArticle, function($location, data, requestUrl) {
		if (data && options.isLogon && syra_site.userProfile) {
			syra_site.userProfile.setCurrentLangCode(data);
		}
		if (options.menuArticle) {
			var segments = syra_url.parse($location.$url = syra_expression.parse(options.menuArticle, $location.$url, data));
			if (options.menu.$target == "inline" && options.menuArticle.isRecordArticle) {
				syra_router.loadRepresentation({
					article: null,
					segments: segments,
					success: function($itemPage) {
						options.menuArticle.loadInlinePage(options.menuArticle, $itemPage, options);
					}
				});
			} else {
				exports.preload = {
					initData: (options.menu.$method != "DELETE") ? data : null,
					openUrl: options.menu.$url
				};
				syra_url.history.load({
					$url: segments.$url,
					target: segments.target
				});
			}
		} else {
			syra_router.loadRepresentation({
				article: null,
				segments: $location.$url,
				success: function($itemPage) {
					options.article.$prototype = $itemPage.$representation.$prototype;
					syra_item.initialize(options.article.page, options.article, {});
					options.article.page.wcpreloadUrl = requestUrl;
					options.article.loadBox(data);
					options.callback && options.callback($itemPage, $location, data);
				}
			});
		}

	});
};

exports.add = function(page) {
	if (exports.isFormUrl(page.openerUrlSegments.$url)) {
		page.isWorkingCopy = true;
		syra_site.activeWorkingCopy = page;
	}
};

exports.remove = function(page) {
	if (page && !page.isFusionPage && page.$isEditMode) {
		if (page == syra_site.activeWorkingCopy) {
			delete syra_site.activeWorkingCopy;
		}
		var $url = page.$prototype ? page.$prototype.$representationUrl : null;
		if (exports.isFormUrl($url)) {
			syra_ajax["delete"]({
				url: $url,
				success: function(data, response) {

				},
				error: function(data, response) {

				}
			});
		}
	}
};

function _onAfterActionMenuExecute(menuItem, $menu, isSuccess, $diagnoses) {
	var page = menuItem.page;
	var isSaved = menuItem.$sourceBind == "$save";
	isSaved && _storage.clear(page);
	var inlinePageHost = page.inlinePageHost;
	if (inlinePageHost) {
		var record = inlinePageHost.record;
		if (isSaved) {
			var $binds = Object.keys(record.$prototype.$properties);
			var $delta = {};
			for (var ii = 0, jj = $binds.length; ii < jj; ii++) {
				var $bind = $binds[ii];
				if (page.dataset[$bind] != undefined) {
					$delta[$bind] = page.dataset[$bind];
				}
			}
			syra_dataset.applyDelta(record.page, record.dataset, $delta);
			record.applyChange($delta);
			setTimeout(function() {
				inlinePageHost.dispose(record.list);
			}, 100);
			return false;
		}
	} else {
		if (isSaved && isSuccess) {
			if (page.dialogWrapper) {
				if (page.dialogWrapper.options.onServerSaved) {
					if (page.dialogWrapper.options.onServerSaved(menuItem, page.dialogWrapper) === false) {
						return false;
					}
				}
			} else {
				var $links = page.$menus.$save.$links;
				var $redirect = $links && ($links.$redirect || $links.$details);
				if ($redirect) {
					var $url = syra_url.formatMenuUrl(page, $redirect);
					$diagnoses = ($diagnoses || []);
					if (page.$menus.$save.$diagnoses) {
						$diagnoses = $diagnoses.concat(page.$menus.$save.$diagnoses);
					}
					(_lastSavingDiagnoses = {})[$url] = $diagnoses;
					syra_router.executeMenu({
						replaceState: true,
						$url: $url
					}, page);
					return false;
				}
			}
		}
	}
	return true;
}

exports.showLastSavingDiagnoses = function(page) {
	if (_lastSavingDiagnoses) {
		var $diagnoses = _lastSavingDiagnoses[page.openerUrlSegments.$url];
		_lastSavingDiagnoses = null;
		$diagnoses && syra_alert.show($diagnoses, page);
	}
};


function _alert(menu, $delta) {
	var $menu = menu.articleParent.$menus[menu.$item.$bind];
	if ($menu) {
		if ($menu.$links) {
			var $binds;
			var $links = $menu.$links || {};
			if ($delta.$links) {
				$binds = Object.keys($delta.$links);
				for (var ii = 0, jj = $binds.length; ii < jj; ii++) {
					var $bind = $binds[ii];
					var $source = $delta.$links[$bind];
					var $target = $links[$bind] = $links[$bind] || {};
					var $props = Object.keys($source);
					for (var kk = 0, mm = $props.length; kk < mm; kk++) {
						var $prop = $props[kk];
						$target[$prop] = $source[$prop];
					}
				}
			}
			$binds = Object.keys($links);
			for (var ii = 0, jj = $binds.length; ii < jj; ii++) {
				var $bind = $binds[ii];
				var $link = $links[$bind];
				$link.$bind = $bind + "-" + menu.id;
				$link.$url = syra_url.formatMenuUrl(menu.articleParent, $link);
			}
			syra_alert.show($delta.$diagnoses, menu.articleParent, {
				$links: $links
			});
		} else {
			$delta.$diagnoses && syra_alert.show($delta.$diagnoses, menu.articleParent);
		}
	}
}


function _getDiagnoses(stack, $diagnoses) {
	var hasError;
	if (stack) {
		for (var ii = 0, jj = stack.length; ii < jj; ii++) {
			var item = stack[ii];
			var $diagnose = syra_site.clone(item.$diagnose);
			if (!hasError) {
				hasError = item.isError();
			}
			if (item.fieldText) {
				$diagnose.$message = item.fieldText + " " + $diagnose.$message;
			}
			$diagnoses.push($diagnose);
		}
	}
	return hasError;
}


exports.onActionExecuted = function(menu, $menu) {
	var isSuccess = true;
	var $diagnoses = [];
	//check diagnoses
	var dgpn = menu.page && menu.page.alertPanel;
	if (dgpn) {
		isSuccess = !_getDiagnoses(dgpn.stack, $diagnoses) && !_getDiagnoses(dgpn.fieldStack, $diagnoses);
	}
	if (menu.$diagnoses) {
		for (var ii = 0; isSuccess && ii < menu.$diagnoses.length; ii++) {
			if (menu.$diagnoses[ii].$severity != "success") {
				isSuccess = false;
			}
		}
	}
	if (menu.articleParent == menu.page) {
		var res = menu.page.onAfterActionMenuExecute ? menu.page.onAfterActionMenuExecute(menu, $menu, isSuccess, $diagnoses) : _onAfterActionMenuExecute(menu, $menu, isSuccess, $diagnoses);
		if (res === false) {
			$menu.$isRequested = true;
			return false;
		}
	}

	$menu.$diagnoses && _alert(menu, $menu);

	if ($menu.$links && $menu.$links.$location) {
		var $url = $menu.$links.$location.$url;
		setTimeout(function() {
			syra_site.unload();
			window.open($url, "_self");
		}, 100);
	}
	if ($menu.$location && menu.$invocationMode == "async") {
		syra_trackers.addTracker({
			$location: $menu.$location,
			$state: $menu.$state,
			title: menu.getTitle(),
			uuid: $menu.$location.split(/'/)[1]
		});
	}
	$menu.$isRequested = true; //avoid processing action for other menu bound to $menu (create menu for sample)
	return true;
};

exports.postAction = function(page, menuItem, target, value) {
	if (page.form_postAction ? page.form_postAction(menuItem, target, value) : true) {
		if (target != page) {
			var list = target.list || target;
			if (list && list.isArrayField) {
				if (value.$delete || value.$create) {
					_postActionArrayType[list.formType](list, menuItem, target, value);
					//list.resizeItem(true);
					return;
				}
			}
		}
		var articleParent = (target.isArticle || target.isSection) ? syra_article.getArticle(target) : target.articleParent; //test for compatibilty with old getArticle
		var sendBag = exports.getSendBag(page);
		var articleSendBag = exports.getSendBag(articleParent);
		var $actions;
		var $bind;
		//var dataset = articleSendBag.dataset || articleSendBag;
		var dataset = articleSendBag;
		if (target.$item && target.$item.$bind) {
			var $properties = dataset.$properties = dataset.$properties || {};
			$properties = $properties[target.$item.$bind] = $properties[target.$item.$bind] || {};
			$actions = $properties.$actions = $properties.$actions || {};
			$bind = "$properties";
		} else {
			$bind = "$actions";
			$actions = dataset.$actions = dataset.$actions || {};
		}
		var $binds = Object.keys(value);
		for (var ii = 0, jj = $binds.length; ii < jj; ii++) {
			$actions[$binds[ii]] = value[$binds[ii]];
		}
		value = dataset[$bind];
		if (articleParent != page) {
			while (articleParent != page) {
				if (articleParent.isRecordArticle) {
					var list = articleParent.list;
					var listSendBag = exports.getSendBag(list);
					if (list.$prototype.$type == "application/x-array") {
						var found = list.findDataRecord(articleParent.$uuid);
						listSendBag.saveRecord(articleParent, $bind, value, found.dataRecordIndex);
						value = listSendBag.dataRecords;
					} else {
						if ($bind) {
							(listSendBag.dataset = listSendBag.dataset || {})[$bind] = value;
						}
						value = listSendBag.dataset;
					}
					$bind = list.$item.$bind;
					articleParent = list.articleParent;
				} else {
					if (articleParent.arrayLevel != "array") {
						if (articleParent.dataset && articleParent.dataset.$uuid) {
							sendBag.$uuid = articleParent.dataset.$uuid;
						}
						sendBag[$bind] = value;
					}
					articleParent = articleParent.articleParent;
				}
			}
			sendBag[$bind] = value;
		}
		if (!page.isFormUpdating) {
			if (page.dataset && page.dataset.$uuid !== undefined) {
				sendBag.$uuid = page.dataset.$uuid;
			}
			_updateWorkingCopy(page, sendBag);
		}
	}
};

var _postActionArrayType = {
	multiArray: function(list, menuItem, target, value) {
		var record = target.isRecordArticle ? target : null;
		if (value.$create) {
			var dataRecords = syra_dataset.ensure(list);
			list.filler.ensureClientDataset(list);
			var options = {
				isCreateAction: true
			};
			if (record) {
				options.$serverIndex = record.$serverIndex;
				options.$recordIndex = record.getRecordIndex();
			} else {
				if (list.dataset.length > 0) {
					options.$serverIndex = list.dataset[list.dataset.length - 1].$serverIndex + 1;
				} else {
					options.$serverIndex = 0;
				}
			}
			var newValue = {
				$uuid: helpers.uuid.generate()
			};
			if (menuItem.$item.$variantItemKey) {
				options.dataRecord = {};
				options.dataRecord[menuItem.$item.$variantItemKey] = newValue;
			} else {
				options.dataRecord = newValue;
			}
			if (record) {
				var found = list.filler.findDataRecord(list, record.$uuid);
				dataRecords.splice(found.dataRecordIndex, 0, options.dataRecord);
			} else {
				dataRecords.splice(options.$serverIndex, 0, options.dataRecord);
			}
			if (list.paging.$startIndex !== undefined && list.clientDataset.length < list.paging.$startIndex) {
				list.paging.ensureClientRange();
			}
			if ((list.paging.$startIndex + (list.paging.$itemsPerPage - 1)) > options.$serverIndex) {
				var newRecord = list.filler.addRecord(list, options);
			}
			//sendbag begin
			var sendBag = exports.getSendBag(list);
			var saved = {
				$uuid: newValue.$uuid
			};
			if (sendBag.dataRecords.length > 0) {
				sendBag._ensureFullDataSet(list);
				if (!sendBag._findDataRecord(saved.$uuid)) {
					sendBag.dataRecords.splice(options.$serverIndex, 0, saved);
				}
			} else {
				saved.$index = options.$serverIndex;
				sendBag.dataRecords.push(saved);
			}
			//sendbag end
			if (list.isClientFetch) {
				options.dataRecord.$serverIndex = options.$serverIndex;
				if (list.clientDataset != list.dataset) {
					list.clientDataset.push(options.dataRecord);
				}
				if (list.clientFetchOptions) {
					list.filler.applyClientFetchOptions(list, list.clientFetchOptions);
				}
				list.paging.setPage(options.$serverIndex);
			} else {
				list.filler.validateDisplay(list);
			}
			exports.updateDelta(list, sendBag.dataRecords);

		} else {
			var sendBag = exports.getSendBag(list);
			var dataRecords = syra_dataset.ensure(list);
			list.filler.ensureClientDataset(list);
			if (record) {
				var clientDataRecordIndex = list.filler.getClientDataSetIndex(list, record.$serverIndex);
				list.records.splice(list.records.indexOf(record), 1);
				var found = list.filler.findDataRecord(list, record.$uuid);
				if (found) {
					dataRecords.splice(found.dataRecordIndex, 1);
					record.$index = found.dataRecordIndex;
				}
				var found = sendBag.ensureRecord(record);
				if (sendBag.dataRecords.length == 1) {
					found.dataRecord.$index = record.$serverIndex;
					found.dataRecord.$isDeleted = true;
				} else {
					sendBag.dataRecords.splice(found.dataRecordIndex, 1);
				}
				delete record.$index;
				if (clientDataRecordIndex >= 0 && list.clientDataset != list.dataset) {
					list.clientDataset.splice(clientDataRecordIndex, 1);
				}
				delete list.recordsMap[record.$uuid];
				list.filler.removeRecord(record, true);
			} else {
				list.clientDataset.splice(0, dataRecords.length);
				dataRecords.splice(0, dataRecords.length);
				sendBag.dataRecords = [];
				list.filler.removeRecords(list);
			}
			if (list.records.length == 0) {
				list.emptyBody();
			} else {
				if (clientDataRecordIndex >= 0 && list.isClientFetch && list.clientDataset.length) {
					list.paging.setPage(list.clientDataset[Math.min(clientDataRecordIndex, list.clientDataset.length - 1)].$serverIndex);
				} else {
					list.filler.validateDisplay(list);
				}
			}
			exports.updateDelta(list, sendBag.dataRecords);
		}
	},
	singleArray: function(list, menuItem, target, value) {
		var sendBag = exports.getSendBag(list);
		var dataRecords = syra_dataset.ensure(list);
		var record = target.isRecordArticle ? target : null;
		if (value.$create) {
			var options = {};
			if (record) {
				options.$serverIndex = record.$serverIndex;
				options.$recordIndex = record.getRecordIndex();
			} else {
				if (list.records.length > 0) {
					options.$serverIndex = list.records[list.records.length - 1].$serverIndex + 1;
				} else {
					options.$serverIndex = 0;
				}
			}
			var targetType = list.$prototype.$item.$type;
			if (menuItem.$item.$variantItemKey) {
				targetType = list.$prototype.$item.$variants[menuItem.$item.$variantItemKey].$type;
			}
			var newValue = (targetType == "application/x-object") ? {
				$uuid: helpers.uuid.generate()
			} : null;
			if (menuItem.$item.$variantItemKey) {
				options.dataRecord = {};
				options.dataRecord[menuItem.$item.$variantItemKey] = newValue;
			} else {
				options.dataRecord = newValue;
			}
			(sendBag.dataRecords = dataRecords).splice(options.$recordIndex || list.records.length, 0, options.dataRecord);
			var record = list.filler.addRecord(list, options);
			list.filler.validateDisplay(list);
			record.scrollToRecord();
		} else {
			if (record) {
				var recordIndex = record.getRecordIndex();
				list.records.splice(recordIndex, 1);
				dataRecords.splice(recordIndex, 1);
				sendBag.dataRecords = dataRecords;
				delete list.recordsMap[record.$uuid];
				list.filler.removeRecord(record, true);
			} else {
				dataRecords.splice(0, dataRecords.length);
				sendBag.dataRecords = [];
				list.filler.removeRecords(list);
			}
			list.filler.validateDisplay(list);

			if (list.records.length == 0) {
				list.emptyBody();
			}
			exports.updateDelta(list, sendBag.dataRecords);
		}
		list.addMenusBar();
	}
};
