// syracuse-ui/lib/controller/officeAddin.js
"use strict";
var _helpers = require('syracuse-core').helpers;

exports.exe = "/msoffice/lib/general/addIn/SyracuseOfficeAddinsSetup.EXE";

exports.installed = function(controller, object, article, title, callback, args) {
	if (object.$confirm && object.$officeAddinSetup) {
		if (_helpers.getCookie("SyracuseOfficeAddin2") === "valid") {
			return true;
		}
	}

	if (object.$url && object.$url.indexOf("SyracuseOfficeAddinsSetup.EXE") > 0) {
		_helpers.setCookie("SyracuseOfficeAddin2", "valid", null, (new Date((new Date().getYear() + 1910), new Date().getMonth(), new Date().getDate())).toUTCString());
	}

	if (object.$confirm) {
		setTimeout(function() {
			syra_site.closePopups();
			syra_alert.ask({
				$isAutoClose: object.$officeAddinSetup ? 10000 : false,
				$title: syra_expression.parse(article, title || syra_local.siteDefaultConfirmTitle),
				$message: syra_expression.parse(article, object.$officeAddinSetup ? syra_local.installOfficeAddin : object.$confirm),
				click: function(response, closureMode) {
					var yes = response && response.$id == "yes";
					if (yes || closureMode == "auto") {
						if (yes && object.$officeAddinSetup) {
							_helpers.setCookie("SyracuseOfficeAddin2", "valid", null, (new Date((new Date().getYear() + 1910), new Date().getMonth(), new Date().getDate())).toUTCString());
						}
						callback.apply(controller, args);
					}
				}
			});
		}, 50);
		return false;
	}
	return true;
};
