// syracuse-ui/lib/controller/soapController.js
"use strict";

exports.isSoapUrl = function($url){
    return $url && ($url.indexOf("soap-stubs") >= 0);
};

exports.openRequestPage = function(httpQuery){
    var soapQuery = document.controller.parseUrl(httpQuery.$url);
    soapQuery.sendRequest({}, function(data, response, requestUrl){
        var $itemPage = {
            $$container: $("<div/>"),
            $isEditMode: true,
            $category: "page",
            $urlParts: soapQuery.$urlParts,
            $representation: data
        };
        $itemPage.$urlParts.$facet = "$request";
        var dialog = document.site.openDialog({
            $itemPage: $itemPage,
            onValidate: function(page){
                if (page.validateFields()) {
                    var invokeMenu = page.$prototype.$links.$invoke;
                    page.formatMenuUrl(invokeMenu, null, function($url){
                        invokeMenu.$url = $url;
                        document.controller.sendRequest(page, {
                            $location: invokeMenu,
                            data: page.sendBag,
                            method: invokeMenu.$method,
                            noDisplayErr: invokeMenu.noDisplayErr || false
                        }, function(data, response){
                            _openResponsePage(invokeMenu.$url, response.data);
                            dialog.close();
                        }, function(err){
                        
                        });
                    });
                }
                return false;
            },
            onClose: function(isCanceled){
            }
        });
    }, function(){
        //onerror
    });
};


function _openResponsePage($url, data){
    var soapQuery = document.controller.parseUrl($url);
    setTimeout(function(){
        var $itemPage = {
            $category: "page",
            $urlParts: soapQuery.$urlParts,
            $representation: data
        };
        $itemPage.$urlParts.$facet = "$response";
        //   document.site.onMainPageChange($itemPage);
        document.controller._openMainPage(soapQuery, $itemPage);
    }, 10);
};
function _openResponseDialogPage($url, data){
    var soapQuery = document.controller.parseUrl($url);
    setTimeout(function(){
        var $itemPage = {
            $$container: $(document.createElement("div")),
            $category: "page",
            $urlParts: soapQuery.$urlParts,
            $representation: data
        };
        $itemPage.$urlParts.$facet = "$response";
        var dialog = document.site.openDialog({
            $itemPage: $itemPage,
            onValidate: function(page){
                return false;
            },
            onClose: function(isCanceled){
            }
        });
    }, 10);
};
