// etna/lib/supervisor/metadata/class.ts
// TODO: eliminate these abnormal dependencies
import { ClassDataChildren, ClassMap } from '../meta/class'
import { Table } from '../meta/table'

export interface ParamDefsMeta {
	CODE: string;
	TYPKEY: number;
	NOPAR: number;
}


export interface ParamFieldMeta {
	FLDCLA: string;
	VALEUR: string;
}

export interface StdMethodMeta {
	ENAMETSTD: number;
	ACTMETSTD: string;
	CODMETSTD: string;
}

export interface MethodParamDefsMeta {
	CODPAR: string;
	TYPINTPAR: number;
	MODPAR: number;
	TYPKEY: number;
	NOPAR: number;
}

export interface MethodMeta {
	CODMET: string;
	FLGOPE: number;
	PARAM_DEFS: MethodParamDefsMeta[];
	DONMET: number;
}

export interface BaseClassMeta {
	CODCLA: string;
	CODACT: string;
	DESCRIPT: string;
	KEYINT: string;
	TYPCLA: number;
	CODREP: string;
	CODTRT: string[];
	PROPERTIES: PropertyMeta[];
	TYPMSKREP: number;
}

export interface ClassMeta extends BaseClassMeta {
	CODCLA: string;
	INTCLA: number;
	MODULE: number;
	PARAM_DEFS: ParamDefsMeta[];
	PARAM_FLDS: ParamFieldMeta[];
	PROPERTIES: PropertyMeta[];
	COLLECTIONS: CollectionMeta[];
	METHODS: MethodMeta[];
	STD_METHODS: StdMethodMeta[];
	LINKS: ClassLinkMeta[];
	MAPS: ClassMap[];
	FILTERS: string[];
	KEYINT: string;
	INDREF: string;
	FLGSEARCH: number;
	FLGBUFFER: number;
	TABREF: string;
	TYPCLA: number;
	CODTRT: string[];
	ACCSTR: string;
	dataChildren: ClassDataChildren;
	codclaPath: string;
	table: Table;
	_proxyName: string;
}

export interface PropertyParamMeta {
	NUMPAR: number;
	TYPPAR: number;
	FLDCLA: string;
	CODFLD: string;
	CODPAR: string;
	VALEUR: string;
	TYPKEY: number;
	TYPINT: number;
	// clean this up
	alias: string;
}

export interface PropertyMeta {
	FLDCLA?: string;
	CODTYP?: string;
	NOLIB?: number;
	INTFLD?: number;
	INTSHTFLD?: number;
	FLGACCGET?: number;
	FLDSEARCH?: number;
	OBLIG?: number;
	LONG?: number;
	FLDGRP?: string;
	CODFLD?: string;
	PARAMS?: PropertyParamMeta[];
	PARAM_FLDS?: PropertyParamMeta[];
	LOBTAB?: string;
	LOBFLD?: string;
	LNKCLA?: string;
	NUMFLD?: number;
	NUMLIG?: number;
	_proxyName?: string;
}

export interface ClassLinkMapMeta {
	PROMAP: string;
	KEYMAP: string;
	VALDEFMAP: string;
}

export interface ClassLinkMeta {
	TYPLNK: number;
	MAPS: ClassLinkMapMeta[];
	CLALNK: string;
	CLELNK: string;
	EXPLNK: string; 
	TABLNK: string;
	[name: string]: any;
}

export interface CollectionMeta {
	CODCOL: string;
	PROCOL: string;
	INTCOL: number;
	MAXCOL: number;
	MINCOL: number;
	FLGAPDCOL: number;
	FLGINSCOL: number;
	FLGSUPCOL: number;
	FLGTRICOL: number;
	ACTGRP: string;
	_proxyName: string;
}
