// etna/lib/supervisor/metadata/representation.ts
import { BaseClassMeta, CollectionMeta } from './class';

export interface ExportedPropertyMeta {
	CODPRO?: string;
	ALIAS?: string;
	ORDPRO?: number;
	INTPRO?: number;
	INTSHTPRO?: number;
	TYPAFFPRO?: number;
	UOMPRO?: string;
	BLCPRO?: string;
	[name: string]: any;
}

export interface LinkParamMeta {
	CODPAR: string;
	VALEUR: string;
	TYPINT: number;
	TYPPAR: number;
	TYPKEY: number;
	PARFIL: number;
	PARENT: number;
	ADRVAL: number;
	// TODO: clean this up
	alias: string;
	value: string;
}

export interface LinkMeta {
	INTLNK: number;
	TYPLNK: number;
	FLGSTDLNK: number;
	CMPLNK: number;
	CODLNK: string;
	ACVLNK: string;
	CLALNK: string;
	REPLNK: string;
	TARLNK: number;
	ATTLNK: number;
	FLGASYLNK: number;
	AFFLNK: number;
	FREELNK: string;
	CODFNCLNK: string;
	ANCLNK: string;
	ENALNK: number;
	REMSTDLNK: string;
	MENLNK: string;
	ORDLNK: number;
	DETLNK: number;
	EDILNK: number;
	SUMLNK: number;
	QRYLNK: number;
	LOKLNK: number;
	PARAMS: LinkParamMeta[];
	ORDTECLNK: number;
	CODOPELNK: string;
	// todo: clean this up
	oCODLNK: string;
	oAFFLNK: number;
	[name: string]: any;
}

export interface FilterMeta {
	OPTFLGCLA: number;
	OPTCOD: string;
	OPTDEF: number;
	OPTOBY: number;
	OPTACT: string;
	OPTLIB: number;
	OPTCND: string;
}

export interface BlockMeta {
	NIVBLC: number;
	CODBLC: string;
	INTBLC: number;
}

export interface SectionMeta {
	NIVSEC: number;
	INTSEC: number;
	BLOCKS: BlockMeta[];
}


export interface MenuMeta {
	CODMENU: string;
	PARMENU: string;
	ACTMENU: string;
	LIBMENU: number;
}

export interface RepresentationMeta extends BaseClassMeta {
	ABRCLA: string;
	INTREP: number;
	EXPORTED_PROPERTIES: ExportedPropertyMeta[];
	COLLECTIONS: CollectionMeta[];
	FILTERS: FilterMeta[];
	LINKS: LinkMeta[];
	SECTIONS: SectionMeta[];
	MENUS: MenuMeta[];
	ENAFAC: number[];
	ENACOM: number[];
	DEFLNKFAC: number[];
	AFFLNKFAC: number[];
	FONCTION: string;
	LNKMENFAC: string[];
	_etag: string;
}
