// syracuse-ui/lib/field/array/capability/recordReorderDDAgent.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function RecordReorderDDAgent() {}

exports.RecordReorderDDAgent = helpers.defineClass(RecordReorderDDAgent, null, {
	start: function(dropItem) {
		syra_dd.ddAgent = this;
		this.reorderRecord = dropItem.reorderRecord;
		this.list = dropItem.list;
		if (this.list.onRecordReorderAgentBeforeStart) {
			this.list.onRecordReorderAgentBeforeStart(this);
		}
		this.boundaryItem = this.boundaryItem || this.list.body;
		this.boundary = this.boundary || this.boundaryItem.getBoundingClientRect();
		this.scrollViewRect = this.list.page.getScrollviewSize();
		this.scrollViewRect.item = this.list.page.scrollview;
		this.scrollViewRect.scrollWidth = this.scrollViewRect.item.scrollWidth;
		this.scrollViewRect.scrollHeight = this.scrollViewRect.item.scrollHeight;
	},
	onDragMouseMove: function(target, event) {
		if (event.target) {
			if (this.boundaryItem.contains(event.target)) {
				var target = event.target;
				while (target && target != this.boundaryItem) {
					if (target.syraRecord) {
						if (this.reorderRecord) {
							this.moveDragImage(event, syra_map[target.syrainout]);
							event.stopPropagation();
							return;
						}
					}
					target = target.parentNode;
				}
			}
		}
		if (!event.target || event.target && event.target.className.indexOf("s-list-drop-cue") < 0) {
			this.moveDragImage(event, null, true);
		}
	},
	onDragMouseUp: function(target, event) {
		if (this.targetRecord) {
			var isAfter = this.$drag.$insert == "insertAfter";
			if (!this.list.saveReorder || this.list.saveReorder(this.reorderRecord, this.targetRecord, isAfter)) {
				var dataRecords = this.list.ensureDataSet();
				var dataRecord = dataRecords[this.reorderRecord.dataset.$serverIndex];
				var $prevReorderedIndex = this.reorderRecord.dataset.$serverIndex;
				dataRecords.splice(this.reorderRecord.dataset.$serverIndex, 1);
				var reorderRecordIndex = this.reorderRecord.getRecordIndex();
				var targetRecordIndex = this.targetRecord.getRecordIndex();
				this.list.records.splice(reorderRecordIndex, 1);
				if (targetRecordIndex > reorderRecordIndex) {
					reorderRecordIndex = isAfter ? targetRecordIndex : (targetRecordIndex - 2);
				} else {
					reorderRecordIndex = isAfter ? targetRecordIndex + 1 : (targetRecordIndex);
				}
				if (this.targetRecord.dataset.$serverIndex > this.reorderRecord.dataset.$serverIndex) {
					this.reorderRecord.dataset.$serverIndex = isAfter ? this.targetRecord.dataset.$serverIndex : (this.targetRecord.dataset.$serverIndex - 2);
				} else {
					this.reorderRecord.dataset.$serverIndex = isAfter ? this.targetRecord.dataset.$serverIndex + 1 : (this.targetRecord.dataset.$serverIndex);
				}
				reorderRecordIndex = Math.max(reorderRecordIndex, 0);
				this.reorderRecord.dataset.$serverIndex = Math.max(this.reorderRecord.dataset.$serverIndex, 0);
				dataRecords.splice(this.reorderRecord.dataset.$serverIndex, 0, dataRecord);
				this.list.records.splice(reorderRecordIndex, 0, this.reorderRecord);
				this.reorderRecord.reorderItem(this.targetRecord, isAfter);
				this.list.validateDisplay();
				var sendBag = this.list.page.ensureSendBag(this.list);
				for (var ii = 0, jj = dataRecords.length; ii < jj; ii++) {
					dataRecords[ii].$serverIndex = ii;
				}
				if (this.list.isSingList) {
					sendBag.dataRecords = dataRecords;
				} else {
					sendBag.dataRecords = [];
					sendBag.ensureRecord(this.reorderRecord);
					sendBag.ensureRecord(this.targetRecord);
				}
				this.list.page.notifyDataChange(this.list, sendBag.dataRecords);
			}
		}
		syra_site.dom.removeChild(this._info);
		syra_site.dom.removeChild(this._dropCue);
	},
	moveDragImage: function(event, targetRecord, isOut) {
		this.$action = null;
		if (targetRecord) {
			if (this.reorderRecord && this.reorderRecord != targetRecord) {
				this.$action = {
					$move: true
				};
				this._ensureDragImage();
				this._info.textContent = syra_local.flMoveItem;
			}
		}
		this.list.page.autoScroll(this.scrollViewRect, event);
		var top = Math.max(event.pageY, this.boundary.top);
		var left = Math.max(event.pageX, this.boundary.left);
		top = Math.min(top, this.boundary.bottom);
		left = Math.min(left, this.boundary.right);

		var inBoundary = (event.pageX == left && event.pageY == top);
		this._ensureDragImage();
		if (!isOut && this.$action) {
			syra_site.dom.toggleClass(this._info, "s-drag-ok", true);
			this.targetRecord = targetRecord;
		} else {
			if (isOut) {
				this._info.textContent = syra_local.flMoveItem.replace("{0}", "move record");
			}
			syra_site.dom.toggleClass(this._info, "s-drag-ok", false);
			this.targetRecord = null;
		}
		this.moveDragCue(event);
		this._info.style.display = "";
		this._info.style.top = top + 15 + "px";
		this._info.style.left = left + 15 + "px";
	},
	moveDragCue: function(event) {
		var $position = null;
		if (event && this.targetRecord) {
			$position = this._calculateDropCuePosition(event);
		}
		if (!this._dropCue) {
			this._dropCue = document.createElement("div");
			this._dropCue.className = "s-list-drop-cue";
			syra_site.layoutSlot.appendChild(this._dropCue);
		}
		this._dropCue.className = "s-list-drop-cue";
		if ($position) {
			this._dropCue.style.display = "block";
			var style = this._dropCue.style;
			style.top = $position.top + "px";
			style.left = $position.left + "px";
			style.width = $position.width;
			style.height = $position.height;
			style.display = "block";
		} else {
			this._dropCue.style.display = "none";
		}
	},
	_calculateDropCuePosition: function(event) {
		if (this.list.onRecordReorderAgentCalculateDragOffset) {
			this.$drag == this.list.onRecordReorderAgentCalculateDragOffset(this, event);
		} else {
			this.$drag = syra_site.dom.getBoundingClientRect(this.targetRecord.dataRow || this.targetRecord.domItem);
			if (this.targetRecord.list.scroller) {
				this.$drag.width = this.targetRecord.list.bodyTableSlot.clientWidth;
			}
		}
		//calculateDropBoundary
		var xmargin = (this.$drag.width * 0.25);
		var ymargin = (this.$drag.height * 0.25);
		var $dropBoundary = {
			left: this.$drag.left + xmargin,
			right: this.$drag.left + this.$drag.width - xmargin,
			top: this.$drag.top + ymargin,
			bottom: this.$drag.top + this.$drag.height - ymargin
		};
		var $position = {
			top: this.$drag.top,
			left: this.$drag.left
		};
		var isAfter = false;
		if (this.reorderRecord.$isVerticalDirection) {
			var bottom = $dropBoundary.bottom;
			if (isAfter = event.pageY > $dropBoundary.bottom) {
				$position.top = $dropBoundary.bottom + ymargin;
			} else {
				$position.top = this.$drag.top;
			}
			$position.height = "0.2em";
			$position.width = this.$drag.width + "px";
		} else {
			isAfter = event.pageX > $dropBoundary.right;
			$position.top -= 3;
			$position.left = this.$drag.left + (isAfter ? (this.$drag.width + 3) : (-3));
			$position.width = "0.2em";
			$position.height = this.$drag.height + 6 + "px";
		}
		this.$drag.$insert = isAfter ? "insertAfter" : "insertBefore";
		return $position;
	},
	_ensureDragImage: function() {
		if (!this._info) {
			this._info = document.createElement("div");
			this._info.style.display = "none";
			this._info.className = "s-list-drag-image";
			syra_site.layoutSlot.appendChild(this._info);
			this._info.textContent = "move record";
		}
	},
	dispose: function() {
		this.list = this.boundaryItem = this.reorderRecord = this.boundary = this._info = this._dropCue = this.$drag = null;
	}
});
