// syracuse-ui/lib/field/array/capability/search.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function Search() {}

exports.Search = helpers.defineClass(Search, null, {
	load: function(list, slot) {
		this.list = list;
		this._item = document.createElement("div");
		this._item.style.display = "none";
		this._item.className = "s-search";
		slot.appendChild(this._item);
		var localize = this.list.localize;
		this._drawItemSearch();
		slot.appendChild(this._item);
		//this._item.style.display = (show !== false) ? "" : "none";
	},
	_drawItemSearch: function() {
		var self = this;
		if (this.list.localize.flSearch) {
			var label = document.createElement("label");
			label.className = "s-search-label";
			this._item.appendChild(label).textContent = this.list.localize.flSearch + ":";
		}

		var link = document.createElement("a");
		link.setAttribute("data-s-picker", "list-search");
		link.setAttribute("href", "#");
		link.syraBind = "$searchCollapse";
		link.className = "s-search-picker-expend";
		this._item.appendChild(link);

		this._search = document.createElement("div");
		this._search.className = "s-search-value";
		this._search.style.display = "none";

		this._itemSearchField = this.list.page.loadNewItem(this._item.appendChild(this._search), {
			$bind: "$search",
			$category: "field",
			$isEditMode: true,
			$isBorderVisible: true,
			$inplace: true,
			$field: {
				$type: "application/x-string",
			}
		}, this.list);

		var div = document.createElement("div");
		div.className = "s-search-link";
		this._search.appendChild(div);
		var link = this._drawSearchLink("Next", "$searchNext");
		link.className = "s-search-next-link";
		div.appendChild(link);
		link = this._drawSearchLink("Previous", "$searchPrevious");
		link.className = "s-search-previous-link";
		div.appendChild(link);
	},
	_drawSearchLink: function(text, $bind) {
		var link = document.createElement("a");
		link.setAttribute("data-s-picker", "list-search");
		link.setAttribute("href", "#");
		link.setAttribute("title", text);
		link.syraBind = $bind;
		return link;
	},
	toggle: function(show) {
		if (this._item) {
			this._item.style.display = (show !== false) ? "" : "none";
		}
	},
	toggleSearch: function(picker) {
		if (this._search) {
			if (this._search.style.display == "none") {
				this._search.style.display = "";
				$(picker).removeClass("s-search-picker-expend").addClass("s-search-picker-collapse");
				this._itemSearchField.focus();
			} else {
				this._search.style.display = "none";
				$(picker).removeClass("s-search-picker-collapse").addClass("s-search-picker-expend");
			}
		}
	},

	clear: function() {
		if (this._itemSearchField) {
			this.list.removeItem(this._itemSearchField);
			delete this._itemSearchField;
		}
		if (this._item) {
			document.site.removeDomChild(this._item);
		}
		this._item = null;
	},
	dispose: function() {
		this.list = this._item = this._itemSearchField = this_search = null;
	}
});
