// syracuse-ui/lib/field/array/capability/search.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function Search() {}

exports.Search = helpers.defineClass(Search, null, {
	load: function(list, slot) {
		this.list = list;
		this._item = document.createElement("div");
		this._item.style.display = "none";
		this._item.className = "s-search";
		slot.appendChild(this._item);
		var localize = this.list.localize;
		this._drawItemSearch();
		slot.appendChild(this._item);
		//this._item.style.display = (show !== false) ? "" : "none";
	},
	_drawItemSearch: function() {
		var self = this;
		if (this.list.localize.flSearch) {
			var label = document.createElement("label");
			label.className = "s-search-label";
			this._item.appendChild(label).textContent = this.list.localize.flSearch + ":";
		}

		var div = document.createElement("div");
		div.className = "s-search-value";
		this._itemSearchField = this.list.page.loadNewItem(this._item.appendChild(div), {
			$bind: "$search",
			$category: "field",
			$isEditMode: true,
			$isBorderVisible: true,
			$inplace: true,
			$field: {
				$type: "application/x-string",
			}
		}, this.list);

		div = document.createElement("div");
		div.className = "s-search-link";
		this._item.appendChild(div);
		var link = this._drawSearchLink("Down", "$searchDown");
		link.className = "s-search-up-link";
		div.appendChild(link);

		link = this._drawSearchLink("Up", "$searchUp");
		link.className = "s-search-down-link";
		div.appendChild(link);
	},
	_drawSearchLink: function(text, $bind) {
		var link = document.createElement("a");
		link.setAttribute("data-s-picker", "list-search");
		link.setAttribute("href", "#");
		link.textContent = text;
		link.syraBind = $bind;
		return link;
	},
	toggle: function(show) {
		if (this._item) {
			this._item.style.display = (show !== false) ? "" : "none";
		}
	},
	clear: function() {
		if (this._itemSearchField) {
			this.list.removeItem(this._itemSearchField);
			delete this._itemSearchField;
		}
		if (this._item) {
			document.site.removeDomChild(this._item);
		}
		this._item = null;
	},
	dispose: function() {
		this.list = this._item = this._itemSearchField = null;
	}
});
