// syracuse-tablet/html/js/init/initStyles.js
"use strict";

/*
 * Module to initialize CSS for used OS/Platform
 *
 */
var utils = require('syracuse-tablet/html/js/helpers/utils');
var globals = require('syracuse-tablet/html/js/helpers/globals');

function _loadCss(file) {
	var ref = document.createElement("link");

	ref.setAttribute("rel", "stylesheet");
	ref.setAttribute("type", "text/css");
	ref.setAttribute("href", file);

	document.getElementsByTagName("head")[0].appendChild(ref);
}

function _getOS() {
	var mobile = (/iphone|ipad|ipod|android|blackberry|mini|windows|palm/i.exec(navigator.userAgent.toLowerCase()));
	var map = {
		iphone: "ios",
		ipad: "ios",
		ipod: "ios",
		android: "android",
		windows: "windows"
	};

	return $.smUrlParam("forceLookAndFeel") || map[mobile || "windows"] || "windows";
}

function _isReleaseVersion() {
	return (document.location.href.indexOf("index.html") > -1);
}

exports.init = function() {
	var deferred = new $.Deferred();
	if (!window.location.pathname.smEndsWith("fdb.html")) {
		var os = _getOS();
		var cssPlatform = "./css/" + os + (_isReleaseVersion() ? ".min" : "") + ".css";
		_loadCss(cssPlatform);
		var cssGeneral = "./css/default.css";
		_loadCss(cssGeneral);
	}
	deferred.resolve();
	globals.setLookAndFeel(os);

	return deferred.promise();
};
