// syracuse-tablet/html/js/init/initStyles.js
"use strict";

/*
 * Module to initialize CSS for used OS/Platform
 *
 */
var globals = require('syracuse-tablet/html/js/helpers/globals');

function _loadCss(file) {
	var ref = document.createElement("link");

	ref.setAttribute("rel", "stylesheet");
	ref.setAttribute("type", "text/css");
	ref.setAttribute("href", file);

	document.getElementsByTagName("head")[0].appendChild(ref);
}

function _getOS() {
	var mobile = (/iphone|ipad|ipod|android|blackberry|mini|windows|palm/i.exec());
	var map = {
		iphone: "ios",
		ipad: "ios",
		ipod: "ios",
		android: "android",
		windows: "windows"
	};

	return $.smUrlParam("forceLookAndFeel") || map[mobile || "windows"] || "windows";
}

function _isSurfaceTablet() {
	navigator.userAgent.toLowerCase().indexOf("tablet pc") > -1 && navigator.userAgent.toLowerCase().indexOf("windows") > -1;
}

function _isReleaseVersion() {
	return (document.location.href.indexOf("index_debug.html") < 0) && (document.location.href.indexOf("index_release.html") < 0);
}

exports.init = function() {
	var deferred = new $.Deferred();
	var os = _getOS();
	var styles;
	//	if (_isSurfaceTablet()) {
	//		styles = "surface";
	//	} else {
	styles = os;
	//	}
	var cssPlatform = "./css/gen/" + styles + (_isReleaseVersion() ? ".min" : "") + ".css";
	_loadCss(cssPlatform);
	var cssGeneral = "./css/default" + (_isReleaseVersion() ? ".min" : "") + ".css";
	_loadCss(cssGeneral);
	deferred.resolve();
	globals.setLookAndFeel(styles);

	return deferred.promise();
};
