// syracuse-tablet/html/js/init/initContext.js
"use strict";

var globals = require('syracuse-tablet/html/js/helpers/globals');
var utils = require('syracuse-tablet/html/js/helpers/utils');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var modalSelectContext = require('syracuse-tablet/html/js/ui/modals/modalSelectContext');
var sdataCommonResources = require('syracuse-tablet/html/js/sdata/sdataCommonResources');
var localStorage = require('syracuse-tablet/html/js/storage/localStorage');
var settings = require('syracuse-tablet/html/js/application/settings');
var native = require('syracuse-tablet/html/js/helpers/native/native');
var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("initCtx");

exports.getLastUsedContext = function(userName) {
	var context;
	try {
		var settingsKey = userName + "_lastContext";
		var contextString = localStorage.getItem(settingsKey);
		if (contextString && contextString.length > 0) {
			context = JSON.parse(contextString);
		}
	} catch (e) {}
	return context;
};

exports.setLastUsedContext = function(userName, context) {
	try {
		var settingsKey = userName + "_lastContext";
		var contextString = JSON.stringify(context);
		localStorage.setItem(settingsKey, contextString);
	} catch (e) {}
};

var _checkContext = function(context) {
	return context && context.language && context.role && context.endpoint ? context : null;
};
/**
 * We assume that
 * 		the user is logged (userCtx and Syracuse session)
 * 		the user profile is available (read after login)
 * fromUserAction	true if it's the use who asks for switching the context (-> dialog)
 * 					false if it's called from an initialization process (after login)
 * contextToRestore	Only if called from initialization process
 * 					Context to restore after login
 * offlineLogin		true if init is called from offline login
 * urlContext		Context provided by url to connect an application/page from email
 * Resolve with
 * 					userContext - Null if login needed
 * 					urlContext	- Null if no url context or if not valid (bad endpoint...)
 */
exports.init = function(fromUserAction, contextToRestore, offlineLogin, urlContext) {
	if (contextToRestore) {
		log && log("Restore context");
		log && log(JSON.stringify(contextToRestore));
	}
	var endpoints = [];
	var roles = [];
	var languages = [];
	var userCtx = contextToRestore ? contextToRestore.userCtx : globals.getUserCtx();
	var userName = userCtx && userCtx.$user;
	var userProfile = contextToRestore ? contextToRestore.userProfile : globals.getUserProfile();
	log && log("User profile");
	log && log(JSON.stringify(userProfile));
	
	// No fail if http status 0
	var noFailIfNoNetwork = true;
	// True if an http status 0 occurred
	var connectionFailedError = false;
	// Settings are read here just so they are put into the globals to make them available later,
	// they are not needed here at all but it's a good place to read them
	return settings.getGlobalSettings().then(function() {
		if (offlineLogin) {
			// No access to common resources because no context is displayed in dialog (only logout)
			return true;
		}
		// Reads endpoint/Language/roles for change profile dialog
		return $.smResolve().then(function() {
			return sdataCommonResources.queryEndpoints(noFailIfNoNetwork);
		}).then(function(rsrcEndpoints, connectionError) {
			connectionFailedError = connectionFailedError || connectionError;
			if (connectionFailedError) {
				endpoints = [];
				return $.smResolve([], true);
			};
			var urlEndpoint = null;
			if (urlContext) {
				var ar = urlContext.endpoint.split(".");
				urlEndpoint = {
					ok: false,
					protocol: ar[0],
					contract: ar[1],
					dataset: ar[2]
				};
			}
			rsrcEndpoints.forEach(function(ep) {
				if (userProfile.endpoints && userProfile.endpoints.some(function(e) {
					return e.$uuid === ep.$uuid;
				})) {
					if (urlEndpoint && !urlEndpoint.ok && ep.protocol == urlEndpoint.protocol && ep.contract == urlEndpoint.contract && ep.dataset == urlEndpoint.dataset) {
						urlEndpoint.ok = true;
					}
					endpoints.push(ep);
				}
			});
			if (urlEndpoint && !urlEndpoint.ok) {
				// Cancel urlContext
				globals.getModal().actionResult(locale.text("urlcontext.bad"), locale.text("urlcontext.badendpoint", [urlContext.endpoint]));
				urlContext = null;
			}
			return sdataCommonResources.queryLocalePreferences(noFailIfNoNetwork);
		}).then(function(rsrcLanguages, connectionError) {
			connectionFailedError = connectionFailedError || connectionError;
			if (connectionFailedError) {
				languages = [];
				return $.smResolve([], true);
			}
			languages = rsrcLanguages;
			return sdataCommonResources.queryRoles(noFailIfNoNetwork);
		}).then(function(rsrcRoles, connectionError) {
			connectionFailedError = connectionFailedError || connectionError;
			if (connectionFailedError) {
				roles = [];
				return true;
			}
			var urlRole = null;
			rsrcRoles.forEach(function(rl) {
				if (userProfile.roles && userProfile.roles.some(function(r) {
					return r.$uuid === rl.$uuid;
				})) {
					if (urlContext && !urlRole && rl.code === urlContext.role) {
						urlRole = rl;
					}
					roles.push(rl);
				}
			});
			if (urlContext) {
				// Set the role - Can be null if current user dosn't have this role (No error - we keep the current user role)
				urlContext.role = urlRole;
			}
			return false;
		}).then(function(connectionError) {
			connectionFailedError = connectionFailedError || connectionError;
			if (connectionFailedError) {
				return true;
			}
			if (endpoints.length < 1 || roles.length < 1) {
				var err = new Error(locale.text("login.roleendpoint.error"));
				err.$redirectPage = "login";
				return $.smReject(err);
			}
			return false;
		});
	}).then(function(connectionError) {
		var currentContext = contextToRestore ? _checkContext(contextToRestore.context) : null;
		currentContext = currentContext ? currentContext : _checkContext(exports.getLastUsedContext(userName));
		if (offlineLogin && !currentContext) {
			// If init is called from offlineLogin login we need a context
			return $.smReject(new Error("offline.nocontext"));
		}
		if (urlContext && urlContext.endpoint && urlContext.role && languages) {
			// Like language is not provided by url context we take the navigator language to avoid display of modalSelectContext
			var navLang = utils.getNavigatorLanguage();
			if (navLang && languages.some(function(l) {
				if (utils.sameLanguage(l.code, navLang)) {
					urlContext.language = l.code;
					return true;
				};
			})) {
				// urlContext can be used as currentContext
				return $.smResolve(urlContext);
			}
		}
		// !!! Dialog is displayed if:
		//  -> called by user (change profile)
		//  -> No context at initialization even if connectionFailedError (only logout button in that case)
		var displayDialog = fromUserAction === true || currentContext == null;
		if (displayDialog) {
			var deferred = $.Deferred();
			// When we open from a link we hide the endpoints because it's given by urlContext
			// We hide the roles if there's a role provided by url context
			// We don't hide the language because it's not given by urlContext
			var opts = {
				hideEndpoints: urlContext != null,
				hideRoles: urlContext != null && urlContext.role != null
			};
			// Ask user to select a context or to logout
			new modalSelectContext.Modal(currentContext, endpoints, roles, languages, opts).show()
				.then(function(selectContext) {
					if (selectContext == null || selectContext.action === "cancel") {
						// Process canceled
						deferred.resolve(null);
					}
					if (selectContext.action === "logout") {
						// Process canceled - goto login
						deferred.reject(new Error("$logout$"));
					} else {
						deferred.resolve(selectContext.action === "ok" ? selectContext : null);
					}
				}).fail(function(e) {
					deferred.reject(e);
				});
			return deferred.promise();
		} else {
			// Restore the context provided or the last context
			return $.smResolve(currentContext);
		}
	}).then(function(currentContext) {
		// Context null means selectContext has been canceled
		if (!currentContext) return null;
		if (urlContext) {
			// Force the endpoint/role with the one provided by url context if any
			// Endpoint is mandatory
			currentContext.endpoint = urlContext.endpoint;
			if (urlContext.role) {
				// Role is not mandatory - If null we keep the current one
				currentContext.role = urlContext.role;
			}
		}
		// Locale for client UI
		locale.setLocale(currentContext.language);
		// Endpoint for storing records locally and to use for gadgets with default endpoint usage
		globals.setEndpoint(currentContext.endpoint);
		// Context for storing records locally
		globals.setUserCtx({
			$user: userCtx.$user,
			$role: currentContext.role,
			$lang: currentContext.language,
		});
		return currentContext;
	}).then(function(currentContext) {
		// Context null means selectContext has been canceled
		if (currentContext == null || offlineLogin === true || connectionFailedError === true) {
			return null;
		}
		if (native.hasCapability("X3WUPApp")) {
			// Context to restore if any
			native.getModule("X3WUPApp").setReconnectContext({
				context: currentContext,
				userProfile: userProfile,
				userCtx: globals.getUserCtx()
			});
		}
		exports.setLastUsedContext(userName, currentContext);
		return _updateServerUserProfile(currentContext, userName, endpoints, roles, languages);
	}).then(function(currentContext) {
		// !! urlContext can be null if not valid
		return $.smResolve(currentContext, urlContext);
	});
};

function _updateServerUserProfile(context, userName, endpoints, roles, languages) {
	return sdataCommonResources.getUserProfile().then(function(up) {
		var change = false;
		var upChanges = {};
		change = endpoints.some(function(ep) {
			if (ep.application + "." + ep.contract + "." + ep.dataset === context.endpoint && ep.$uuid !== up.selectedEndpoint.$uuid) {
				upChanges.selectedEndpoint = {
					$uuid: ep.$uuid
				};
				return true;
			}
			return false;
		}) || change;
		change = roles.forEach(function(role) {
			if (role.code === context.role && role.$uuid !== up.selectedRole.$uuid) {
				upChanges.selectedRole = {
					$uuid: role.$uuid
				};
				return true;
			}
			return false;
		}) || change;
		languages.forEach(function(lang) {
			if (lang.code === context.languages && lang.$uuid !== up.selectedLocale.$uuid) {
				upChanges.selectedLocale = {
					$uuid: lang.$uuid
				};
				return true;
			}
			return false;
		}) || change;

		if (change) {
			return sdataCommonResources.updateUserProfile(upChanges)
				.then(function() {
					return context;
				});
		} else {
			return $.smResolve(context);
		}
	});
}
