// syracuse-tablet/html/js/ui/modals/modalConfirm.js
"use strict";

var globals = require('syracuse-tablet/html/js/helpers/globals');
var utils = require('syracuse-tablet/html/js/helpers/utils');
var modalDialog = require('syracuse-tablet/html/js/ui/modals/modalDialog');
var locale = require('syracuse-tablet/html/js/helpers/locale');

var _templateAll = {
	header: '\
		<div class="modal-title">{{title}}</div>',
	content: '\
			<div class="s-m-modal-content">\
				{{message}}\
			</div>',
	footer: '\
		<button type="button" class="btn btn-default" data-action="yes">{{yes}}</button> \
		<button type="button" class="btn btn-default" data-action="no">{{no}}</button>'
};

var _checkMsg = function(msg) {
	return msg ? msg.replace(/\n/g, "<br>") : "";
};

var _execCb = function(cb, msg, data) {
	if (!cb || arguments.length == 0) return;
	var cb = arguments[0];
	var args = Array.prototype.slice.call(arguments, 1);
	cb.apply(null, args);
};

var _Klass = utils.defineClass(
	function ModalConfirm(message, action, cb, isSimpleConfirm) {
		var self = this;
		modalDialog.ModalBase.call(self);
		// Init params
		self.message = message;
		self.action = action;
		self.cb = cb;
		self.isSimpleConfirm = isSimpleConfirm;
	}, modalDialog.ModalBase, {
		_getTemplates: function() {
			var self = this;
			return {
				std: _templateAll
			};
		},
		getDisplayFlags: function() {
			var self = this;
			var displayFlags = self.displayFlags || {};
			if (self.deviceType === "tablet") {
				displayFlags.modalClass = "confirm tablet";
			} else {
				displayFlags.modalClass = "confirm mobile";
			}
			return displayFlags;
		},
		_getDataContext: function() {
			var self = this;
			var ctx = {};
			ctx.title = self.isSimpleConfirm ? null : locale.text("modal.confirm.ttl");
			// different message depending on whether the calling modal.js method is confirm() or simpleConfirm()
			ctx.message = self.action ? _checkMsg(locale.text("modal.confirm.msg", [self.action])) : _checkMsg(self.message);
			ctx.yes = locale.text("yes");
			ctx.no = locale.text("no");
			return ctx;
		},
		_onAction: function(action, param) {
			var self = this;
			_execCb(self.cb, action, self.$$elmt, param);
			// close on actions
			return true;
		}
	}
);
exports.Modal = _Klass;
