// syracuse-tablet/html/js/ui/modals/modalNewLayout.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var modalDialog = require('syracuse-tablet/html/js/ui/modals/modalDialog');

var _templateAll = {
	header: '\
		<buttons class="pull-right"> \
			<button type="button" class="btn btn-default s-m-modal-btn-validate" data-action="$validate"></button> \
			<button type="button" class="btn btn-default s-m-modal-btn-cancel" data-action="$cancel"></button> \
		</buttons> \
		<div class="modal-title">{{title}}</div> \
			',
	content: ' \
			<div class="s-m-scroll-wrapper s-m-scroll-wrapper-v">\
				<div class="alert alert-warning" role="alert"> \
					{{warning}} \
				</div> \
				<div class="form-group" id="dir"> \
					<label>{{lab_dir}}</label> \
        			<div class="btn-group btn-group-vertical btn-group-justified" data-toggle="buttons"> \
						<label class="btn btn-primary {{#if horizontal}}active{{/if}}"> \
							<input type="radio" id="horizontal" name="scroll" value="horizontal" {{#if horizontal}}checked{{/if}}/>{{lab_horizontal}}\
						</label> \
						<label class="btn btn-primary {{#unless horizontal}}active{{/unless}}"> \
							<input type="radio" id="vertical" name="scroll" value="vertical" {{#unless horizontal}}checked{{/unless}}/>{{lab_vertical}}\
						</label> \
					</div> \
				</div> \
				<div class="form-group" id="tpl"> \
					<label>{{lab_tpl}}</label> \
					<div class="btn-group btn-group-vertical btn-group-justified" data-toggle="buttons"> \
						<label class="btn btn-primary {{#if template}}active{{/if}}"> \
							<input type="radio" id="template" name="empty" value="true" {{#if template}}checked{{/if}}/>{{lab_template}}\
						</label> \
						<label class="btn btn-primary  {{#unless template}}active{{/unless}}"> \
							<input type="radio" id="empty" name="empty" value="false" {{#unless template}}checked{{/unless}}/>{{lab_empty}}\
						</label> \
					</div> \
				</div> \
			</div>',
	footer: '',
};

var _Klass = utils.defineClass(
	function ModalNewLayout(_defaults) {
		var self = this;
		modalDialog.ModalBase.call(self);

		// Init params
		self._defaults = _defaults;
		// Result
		self.result = null;
	}, modalDialog.ModalBase, {
		_getTemplates: function() {
			var self = this;
			return {
				std: _templateAll
			};
		},

		getDisplayFlags: function() {
			var self = this;
			var displayFlags = self.displayFlags || {};
			if (self.deviceType === "tablet") {
				displayFlags.modalClass = "half_right";
			} else {
				displayFlags.modalClass = "full";
			}
			return displayFlags;
		},

		_getDataContext: function() {
			var self = this;
			var data = {
				label_validate: locale.text("modal.ok"),
				label_cancel: locale.text("modal.cancel"),
				title: locale.text("modal.newlayout.title"),

				warning: locale.text("modal.newLayout.warning"),
				lab_horizontal: locale.text("modal.newLayout.horizontal"),
				lab_vertical: locale.text("modal.newLayout.vertical"),
				lab_template: locale.text("modal.newLayout.template"),
				lab_empty: locale.text("modal.newLayout.empty"),
				lab_dir: locale.text("modal.newLayout.dir"),
				lab_tpl: locale.text("modal.newLayout.tpl"),

				horizontal: self._defaults.scroll === "horizontal",
				template: self._defaults.template
			};

			return data;
		},

		_onValidate: function() {
			var self = this;
			var dir = self.$$elmt.find('#dir label.active input').val();
			var tpl = self.$$elmt.find('#tpl label.active input').val();

			self.result = {
				scroll: dir,
				template: tpl === "true"
			};
		},
		_onCancel: function() {
			var self = this;
			self.result = null;
		},

		_getResult: function() {
			var self = this;
			return self.result;
		}
	}
);

exports.Modal = _Klass;
