// syracuse-tablet/html/js/ui/modals/modalMultiButtons.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var modalDialog = require('syracuse-tablet/html/js/ui/modals/modalDialog');
var siteLayout = require('syracuse-tablet/html/js/ui/siteLayout');

var modules = require('syracuse-tablet/html/js/common/modules');

var _templateAll = {
	header: '\
		{{#if title}}<div class="modal-title">{{title}}</div>{{/if}}',
	content: '\
		<div class="s-m-modal-content">\
			<p>{{{message}}}</p>\
		</div>',
	footer: '\
	{{#each buttons}}\
		<button type="button" {{#if isHidden}}style="display:none"{{/if}} class="btn btn-default {{#if smallBtns}}btn-sm{{/if}}{{#if isDisabled}}disabled{{/if}}" data-action="{{action}}">{{label}}</button>\
	{{/each}}'
};
/**
 * Displays a dialog yes/no - yes/no/cancel ... according to opts
 * opts : {
 * 		buttons:[{action:"yes", label:"yes", isHidden: false, , isDisabled: false}],
 * 		defaultAction: "no"
 * }
 */
var _Klass = utils.defineClass(
	function ModalMultiButtons(title, message, opts) {
		modalDialog.ModalBase.call(this);
		this._dataContext = $.extend({
			title: title,
			message: modalDialog.checkMsg(message || "")
		}, opts);
		this.selectedAction = opts.defaultAction || null;
		var sl = modules.get("siteLayout");
		this._dataContext.buttons.forEach(function(b) {
			b.smallBtns = sl.getDeviceType() === "smartphone" && sl.getPageOrientation() === "portrait";
			b.label = b.label.toUpperCase();
		});
	}, modalDialog.ModalBase, {
		_getTemplates: function() {
			return {
				std: _templateAll
			};
		},
		_onShow: function(evt) {
			var self = this;
			self._css(siteLayout.getPageOrientation(), siteLayout.getDeviceType())
		},
		getDisplayFlags: function() {
			var res = this.displayFlags || {};
			if (this.deviceType === "tablet") {
				res.modalClass = "multibuttons tablet";
			} else {
				res.modalClass = "multibuttons mobile";
			}
			return res;
		},
		_getDataContext: function() {
			return this._dataContext;
		},
		_onAction: function(action, param) {
			this.selectedAction = action;
			// close on actions
			return true;
		},
		_getResult: function() {
			return this.selectedAction;
		},

		_checkButtonSize: function() {
			var sl = modules.get("siteLayout");
			this.$$elmt.find("button").toggleClass("btn-sm", sl.getDeviceType() === "smartphone" && sl.getPageOrientation() === "portrait");
		},

		notifMainLayoutChanged: function(info, orientation, deviceType) {
			this._checkButtonSize();
			if (info.orientationChanged) {
				this._css(orientation, deviceType)
			}
		},
		_css: function(orientation, deviceType) {
			if (deviceType !== "smartphone") return;
			var $$modalDialog = $(".modal-dialog", this.$$elmt);
			if (orientation === "portrait") {
				$$modalDialog.css("margin-top", "50%")
			} else {
				$$modalDialog.css("margin-top", "")
			}
		},
	}
);
exports.Modal = _Klass;
