// syracuse-tablet/html/js/ui/modals/modalPinAndLiveTile.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var modalDialog = require('syracuse-tablet/html/js/ui/modals/modalDialog');
var locale = require('syracuse-tablet/html/js/helpers/locale');

var _templateAll = {
	header: '\
		<buttons class="pull-right"> \
			<button type="button" class="btn btn-default s-m-modal-btn-validate" data-action="$validate"></button> \
			<button type="button" class="btn btn-default s-m-modal-btn-cancel" data-action="$cancel"></button> \
		</buttons> \
		<div class="modal-title">{{subtitle}}</div> \
			',
	content: ' \
			<div class=""> \
				<div class="form-group"> \
					<table class="s-m-pinandlivetile">\
					{{#each items}}\
					<tr><td>{{this.title}}</td>\
					<td><input id="{{this.id}}" type="checkbox" data-value="{{this.value}}"></td><tr>\
					{{/each}}\
					</tr> \
					</table> \
				</div> \
			</div>',
	footer: '',
};
var _items = [{
	id: "pinwelcome",
	value: false
}, {
	id: "livetile",
	value: false
}];
/*
ctx: {
	livetile: false,
	pageTitle: "Customer Dashboard",
	pinwelcome: true
}
*/
var _Klass = utils.defineClass(
	function ModalPinAndLiveTile(ctx) {
		this.context = $.extend(true, {}, ctx);
		modalDialog.ModalBase.call(this);
		this.result = null;
	}, modalDialog.ModalBase, {
		_getTemplates: function() {
			return {
				std: _templateAll
			};
		},
		getDisplayFlags: function() {
			var self = this;
			var displayFlags = self.displayFlags || {};
			if (self.deviceType === "tablet") {
				displayFlags.modalClass = "pinpage half_right";
			} else {
				displayFlags.modalClass = "pinpage full";
			}
			return displayFlags;
		},
		_getDataContext: function() {
			var self = this;
			_items.forEach(function(item) {
				item.value = false;
				item.title = locale.text(item.id + ".page.add");
			});
			var data = {};
			data.items = _items;
			data.title = this.context.pageTitle || "";
			data.subtitle = locale.text("pin.page.subtitle");
			return data;
		},
		_onShow: function() {
			var self = this;
			modalDialog.ModalBase.prototype._onShow.call(this);
			_items.forEach(function(item) {
				var $$item = self.$$elmt.find("#" + item.id);
				self.context[item.id] = self.context[item.id] || false;
				$$item.bootstrapSwitch("onText", locale.text(item.id + ".page.label.onText"));
				$$item.bootstrapSwitch("offText", locale.text(item.id + ".page.label.offText"));
				$$item.bootstrapSwitch("size", "small");
				$$item.bootstrapSwitch("onColor", "default");
				$$item.bootstrapSwitch("offColor", "default");
				$$item.bootstrapSwitch('state', self.context[item.id]);
			});
		},
		_onValidate: function() {
			var content = null,
				self = this;
			_items.forEach(function(item) {
				var $$item = self.$$elmt.find("#" + item.id);
				var value = $$item.bootstrapSwitch('state');
				if (value != self.context[item.id]) {
					content = content || {};
					content["toggle" + item.id] = "change";
				}
			});
			this.result = {};
			this.result.action = content ? "yes" : "no";
			this.result.content = content;
			console.log("_onValidate", this.result);
		},
		_onCancel: function() {
			this.result = {};
			this.result.action = "no";
			this.result.content = null;
		},
		_getResult: function() {
			return this.result;
		}
	}
);

exports.Modal = _Klass;
