// syracuse-tablet/html/js/ui/modals/modalYesNoCancel.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var modalDialog = require('syracuse-tablet/html/js/ui/modals/modalDialog');
var locale = require('syracuse-tablet/html/js/helpers/locale');

var _templateAll = {
	header: '\
		{{#if title}}<div class="modal-title">{{title}}</div>{{/if}}',
	content: '\
		<div class="s-m-modal-content">\
			<p>{{message}}</p>\
		</div>',
	footer: '\
		{{#if showCancel}}<button type="button" class="btn btn-default" data-action="cancel">{{cancel}}</button>{{/if}}\
		{{#if showNo}}<button type="button" class="btn btn-default" data-action="no">{{no}}</button>{{/if}}\
		{{#if showYes}}<button type="button" class="btn btn-default" data-action="yes">{{yes}}</button>{{/if}}'
};
/**
 * Displays a dialog yes/no - yes/no/cancel ... according to opts
 */
var _Klass = utils.defineClass(
	function ModalYesNoCancel(title, message, opts) {
		modalDialog.ModalBase.call(this);
		this._dataContext = $.extend({
			showYes: true,
			showNo: true,
			showCancel: true,
			yes: locale.text("yes"),
			no: locale.text("no"),
			cancel: locale.text("cancel"),
			title: title,
			message: message || "",
			selectedAction: "cancel"
		}, opts);
	}, modalDialog.ModalBase, {
		_getTemplates: function() {
			return {
				std: _templateAll
			};
		},
		getDisplayFlags: function() {
			var res = this.displayFlags || {};
			res.modalClass = "yesnocancel " + this.deviceType;
			return res;
		},
		_getDataContext: function() {
			return this._dataContext;
		},
		_onAction: function(action, param) {
			this.selectedAction = action;
			// close on actions
			return true;
		},
		_getResult: function() {
			return this.selectedAction;
		}
	}
);
exports.Modal = _Klass;
