// syracuse-tablet/html/js/ui/modals/modalShowLegal.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var modalDialog = require('syracuse-tablet/html/js/ui/modals/modalDialog');

var _templateAll = {
	header: '\
			<buttons> \
				<button type="button" class="btn btn-default s-m-modal-btn-cancel pull-right" data-action="$cancel"></button> \
			</buttons> \
			<div class="modal-title">{{title}}</div> \
			',
	content: ' \
			<div class="s-m-scroll-wrapper s-m-scroll-wrapper-v">\
			<div class="s-m-scroll-element"> \
				<div class="form-group"> \
					<label>{{label_update}} {{update}}</label> \
				</div> \
				<div class="form-group"> \
					<label>{{text1}}</label> \
					<label>{{text2}}</label> \
					<label>{{text3}}</label> \
					<label>{{text4}}</label> \
					<label>{{text5}}</label> \
					<label>{{text6}}</label> \
				</div> \
			</div> \
			</div>',
	footer: ''
};

var _Klass = utils.defineClass(
	function ContextSelectionDialog() {
		var self = this;
		modalDialog.ModalBase.call(self);
	}, modalDialog.ModalBase, {
		_getTemplates: function() {
			var self = this;
			return {
				std: _templateAll
			};
		},

		getDisplayFlags: function() {
			var self = this;
			var displayFlags = self.displayFlags || {};
			if (self.deviceType === "tablet") {
				displayFlags.modalClass = "config-tile half_right";
			} else {
				displayFlags.modalClass = "config-tile full";
			}
			return displayFlags;
		},

		_getDataContext: function() {
			var self = this;
			var data = {
				title: locale.text("modal.legal.title"),
				label_update: locale.text("modal.legal.update"),
				update: "8",
				text1: locale.text("modal.legal.text1"),
				text2: locale.text("modal.legal.text2"),
				text3: locale.text("modal.legal.text3"),
				text4: locale.text("modal.legal.text4"),
				text5: locale.text("modal.legal.text5"),
				text6: locale.text("modal.legal.text6")
			};

			return data;
		},

		_onValidate: function() {},

		_onCancel: function() {},

		_getResult: function() {}
	}
);

exports.Modal = _Klass;
