// syracuse-tablet/html/js/ui/modals/modalTranslationPage.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var modalDialog = require('syracuse-tablet/html/js/ui/modals/modalDialog');
var globals = require('syracuse-tablet/html/js/helpers/globals');

var _templateAll = {
	header: '\
		<buttons class="pull-right"> \
			<button type="button" class="btn btn-default s-m-modal-btn-validate" data-action="$validate"></button> \
			<button type="button" class="btn btn-default s-m-modal-btn-cancel" data-action="$cancel"></button> \
		</buttons> \
		<div class="modal-title">{{title}}</div> \
			',
	content: ' \
			<div class="s-m-scroll-wrapper s-m-scroll-wrapper-v">\
				<div class="form-group"> \
					<table class="s-m-translation" >\
					<tr>\
					<th>{{currentLang.title}}</th><th>{{translateLang.title}}</th>\
					</tr>\
					{{#each localizations}}\
					<tr>\
					<td>{{this.value}}</td><td><input type="text" id="{{this.key}}" value="{{this.valueTranslate}}"></td>\
					</tr>\
					{{/each}}\
					</table>\
				</div> \
			</div>',

	footer: '',
};

var _Klass = utils.defineClass(
	function ModalTranslationPage(ctx) {
		this.context = $.extend(true, {}, ctx);
		modalDialog.ModalBase.call(this);
		this.result = null;
	}, modalDialog.ModalBase, {
		_getTemplates: function() {
			return {
				std: _templateAll
			};
		},
		_getDataContext: function() {
			var localizations = [],
				$localization = this.context.$localization,
				currentLang = this.context.currentLang,
				translateLang = this.context.translateLang;
			if (!$localization[currentLang.value]) return [];
			Object.keys($localization[currentLang.value]).forEach(function(key) {
				if ($localization[currentLang.value][key]) {
					localizations.push({
						'key': key,
						'value': $localization[currentLang.value][key],
						'valueTranslate': ($localization[translateLang.value] && $localization[translateLang.value][key]) ? $localization[translateLang.value][key] : "",
					});
				}
			});
			var data = {};
			data.localizations = localizations;
			data.currentLang = currentLang;
			data.translateLang = translateLang;
			data.title = this.context.pageTitle;
			return data;
		},
		_onShow: function() {
			modalDialog.ModalBase.prototype._onShow.call(this);
		},
		_onValidate: function() {
			var currentLang = this.context.currentLang.value,
				translateLang = this.context.translateLang.value,
				localization = this.context.$localization,
				value, res;

			Object.keys(localization[currentLang]).forEach(function(key) {
				value = $("input[id='" + key + "'").val();
				if (localization[translateLang] && localization[translateLang][key] !== "") { //keep existing value
					if (!res) res = {};
					res[key] = localization[translateLang][key];
				}
				if ((!localization[translateLang] && value !== "") || //no localization and at less a value filled
					(localization[translateLang] &&
						((localization[translateLang][key] && value != localization[translateLang][key]) || //key but value dif
							(!localization[translateLang][key] && value)))) { //no key new value
					if (!res) res = {};
					res[key] = value;
				}
			});
			if (res) {
				localization[translateLang] = res;
				this.result = localization;
			}
		},
		_onCancel: function() {
			this.result = null;
		},
		_getResult: function() {
			return this.result;
		}
	}
);

exports.Modal = _Klass;
