// syracuse-tablet/html/js/ui/modals/modalChooseWelcomeDashboard.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var Base = require('syracuse-tablet/html/js/ui/modals/modalChoose').ModalChoose;

var _Klass = utils.defineClass(
	function ModalChooseWelcomeDashboard(dataList) {
		var self = this;
		Base.call(self, dataList.views);
		self.result = null;
	}, Base, {
		_getDataContext: function() {
			var self = this;
			var datactx = [];
			this.dataList.forEach(function(a) {
				datactx.push({
					title: a.$description,
					info: a.$factoryOwner,
					selected: (a.$selected ? "active" : ""),
					dataParams: a.$uuid
				});
				a.key = a.$uuid;
			});
			return Base.prototype._getDataContext.call(this, datactx, {
				title: locale.text("welcome.dashboard.choose.template")
			});
		},
		_onCancel: function() {
			this.result = null
		},
		_onAction: function(action, param) {
			if (action === "clickLine") {
				this.result = {
					selectedView: param,
				}
			}
			return true;
		},
		_getResult: function() {
			return this.result;
		}
	}
);

exports.Modal = _Klass;
