// syracuse-tablet/html/js/ui/modals/modalChooseApps.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var Base = require('syracuse-tablet/html/js/ui/modals/modalChoose').ModalChoose;

var _Klass = utils.defineClass(
	function ModalChooseApps(dataList) {
		Base.call(this, dataList);
	}, Base, {
		_getDataContext: function() {
			var datactx = [];
			this.dataList.forEach(function(a) {
				datactx.push({
					title: a.title,
					info: a.description,
					dataParams: a.applicationName
				});
				a.key = a.applicationName;
			});
			return Base.prototype._getDataContext.call(this, datactx, {
				buttonValidate: true,
				title: locale.text("welcome.select.app.title")
			});
		},
		_onAction: function(action, param) {
			var self = this;
			if (action === "clickLine") {
				var $$item = self.$$elmt.find('[data-params="' + param + '"]');
				if (self.selected[param]) {
					$$item.removeClass("active");
					delete self.selected[param];
				} else {
					$$item.addClass("active");
					self.selected[param] = self._selectedElmt(param);
				}
			}
			// dont close on actions
			return false;
		},
		_selectedElmt: function(key) {
			var match;
			this.dataList.some(function(elmt) {
				if (elmt.key === key) {
					match = elmt;
					return true;
				}
			});
			return match;
		},
		_getResult: function() {
			var self = this;
			return Object.keys(self.selected).map(function(key) {
				return self.selected[key];
			});
		},
		_onCancel: function() {
			this.selected = {};
		},
	}
);

exports.Modal = _Klass;
