// syracuse-tablet/html/js/ui/modals/modalChooseDashboardGroup.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var modalDialog = require('syracuse-tablet/html/js/ui/modals/modalDialog');

var _templateAll = {
	header: '\
			<buttons class="pull-right"> \
				<button type="button" class="btn btn-default s-m-modal-btn-cancel" data-action="$cancel"></button> \
			</buttons> \
			<div class="modal-title">{{title}}</div> \
			',
	content: ' \
			<div class="s-m-scroll-wrapper s-m-scroll-wrapper-v">\
				<div class="form-group s-m-scroll-element"> \
					<div class="list-group">\
						{{#each groups}} \
							<a draggable="false" href="#" class="list-group-item" data-action="clickGroup" data-params="{{id}}">\
								<h4 class="list-group-item-heading">{{title}}</h4>\
							</a>\
						{{/each}} \
					</div> \
				</div> \
			</div>',
	footer: ''
};

var _Klass = utils.defineClass(
	function ModalChooseDashboardGroup(groups) {
		var self = this;
		modalDialog.ModalBase.call(self);
		groups = groups || []
			// Init params
		self.groups = [];
		groups.forEach(function(group) {
			self.groups.push({
				"title": group.title,
				"id": group.id,
			});
		});

		// Result
		self.selectedGroup = {};
	}, modalDialog.ModalBase, {
		_getTemplates: function() {
			var self = this;
			return {
				std: _templateAll
			};
		},

		getDisplayFlags: function() {
			var self = this;
			var displayFlags = self.displayFlags || {};
			if (self.deviceType === "tablet") {
				displayFlags.modalClass = "half_right";
			} else {
				displayFlags.modalClass = "full";
			}
			return displayFlags;
		},

		_getDataContext: function() {
			var self = this;
			return {
				title: locale.text("welcome.move.tile.title"),
				groups: self.groups
			};
		},
		_onCancel: function() {
			var self = this;
			self.selectedgroup = null;
		},

		_onAction: function(action, param) {
			var self = this;
			if (action === "clickGroup") {
				self.selectedgroup = param;
			}
			// close on actions
			return true;
		},

		_getResult: function() {
			var self = this;
			return self.selectedgroup;
		}
	}
);

exports.Modal = _Klass;
