// syracuse-tablet/html/js/ui/modals/modalDiagnoses.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var modalDialog = require('syracuse-tablet/html/js/ui/modals/modalDialog');
var scroller = require('syracuse-tablet/html/js/helpers/scroller/scrollerHelpers');
var uiRect = require('syracuse-tablet/html/js/ui/rect');
var hammerScroller = require('syracuse-tablet/html/js/helpers/scroller/hammerScroller');
var siteLayout = require('syracuse-tablet/html/js/ui/siteLayout');

var _templateAll = {
	header: '\
		<div class="modal-title {{severity}}">{{title}}</div>',
	content: '\
		<div class="s-m-modal-diag-content">\
			{{#each diags}} \
				<message id="s-m-msg-block-{{id}}" role="alert" style="display:{{this.display}}"> \
					<div style="white-space: pre-line; user-select: initial;">{{this.message}}</div> \
					{{#if stack}} \
						<section>\
							<div class="s-m-scroll-wrapper s-m-scroll-wrapper-v">\
								<stack name="s-m-scroll-element stack" class="s-m-msg-stack-content"> \
									{{{this.stack}}} \
								</stack> \
							</div> \
						</section>\
					{{/if}} \
				</message> \
			{{/each}}\
		</div>\
		<div class="s-m-modal-diag-goto">\
			{{#each diags}}\
				<a id="s-m-msg-goto-{{id}}" data-action="goto" data-params="{{id}}">\
					<i class="fa fa-circle-o" data-action="goto" data-params="{{id}}"></i>\
				</a>\
			{{/each}}\
		</div>',
	footer: '\
		<buttons>\
		{{#if showmore}}\
			<button type="button" id="s-m-msg-stacks" class="btn btn-default" data-action="more">{{stackShow}}</button>\
		{{/if}}\
			<button type="button" class="btn btn-default" data-action="$cancel">{{close}}</button>\
		</buttons>\
		'
};

var _Klass = utils.defineClass(
	function ModalDiagnoses(title, diag) {
		var self = this;
		//Simulate mor diag
		/*var diagAdd={
				$message:"test message2",
				$severity:"error",
				$stackTrace:"Stack trace message 2 ↵ligne ↵ligne ↵linge",
		}
		diag.$diagnoses.push(diagAdd);
		 diagAdd={
					$message:"test message 3",
					$severity:"error",
					$stackTrace:"Stack trace message 3 ↵ligne ↵ligne ↵linge",
			}
		diag.data.$diagnoses.push(diagAdd)
		*/
		modalDialog.ModalBase.call(self);
		// Init params
		self.title = title;
		self.diag = diag;
		self.diagId = 0;
		self.curMsg = 0;
	}, modalDialog.ModalBase, {
		destroy: function() {
			var self = this;
			// destroy hammer component
			this._destroySwiper();
			// destroy gestureMgr (used for vertical swipe within the stack)
			if (this._gestureMgrMap && !$.isEmptyObject(this._gestureMgrMap)) {
				$.each(this._gestureMgrMap, function(key, gestureMgr) {
					gestureMgr.destroy();
				});
				this._gestureMgrMap = null;
			}
			modalDialog.ModalBase.prototype.destroy.call(self);
		},
		_getTemplates: function() {
			var self = this;
			return {
				std: _templateAll
			};
		},
		getDisplayFlags: function() {
			var self = this;
			var displayFlags = self.displayFlags || {};
			if (self.deviceType === "tablet") {
				displayFlags.modalClass = "diagnoses tablet";
			} else {
				displayFlags.modalClass = "diagnoses mobile";
			}
			return displayFlags;
		},
		_getDataContext: function() {
			var self = this;
			var ctx = {
				close: locale.text("modal.btn.close").toUpperCase() || "CLOSE",
				stackShow: locale.text("modal.diagnoses.btnstack_show").toUpperCase(),
				severity: "info",
				diags: [],
				showmore: false
			};
			self.diag.$diagnoses.forEach(function(d) {
				var sev;
				if (d.$severity === "success" || d.$severity === "info" || d.$severity === "warning") {
					sev = d.$severity;
				} else {
					sev = "danger";
					ctx.severity = "error";
				}
				var ctxDiag = {
					id: self.diagId,
					display: self.diagId === 0 ? "block" : "none",
					message: d.$message,
					severity: sev,
					severityLabel: locale.text("severity." + sev)
				}
				if (d.$stackTrace) {
					var a = d.$stackTrace.split("\n");
					if (a.length > 20) {
						a.length = 20;
						a.push("...");
					}
					ctxDiag.stack = a.join("<br>");
					ctx.showmore = true;
				}
				ctx.diags.push(ctxDiag);
				self.diagId++;
			});

			self.title = self.title || (ctx.severity === "error" ? locale.text("modal.error.ttl") : locale.text("modal.info.ttl"));
			ctx.title = self.title;
			return ctx;
		},
		_onAction: function(action, param) {
			var self = this;
			switch (action) {
				case "next":
					self._next();
					break;
				case "prev":
					self._prev();
					break;
				case "more":
					self._more();
					break;
				case "goto":
					self._goTo(param);
					break;
			}

			// dont close on actions
			return false;
		},
		_goTo: function(id) {
			var self = this;
			for (var ii = 0, jj = self.diagId; ii < jj; ii++) {
				$("#s-m-msg-block-" + ii, self.$$elmt).css("display", ii == id ? "block" : "none");
			}
			self.curMsg = parseInt(id, 10);
			self._updateButtons();
		},
		_next: function() {
			var self = this;
			if (self.curMsg < (self.diagId - 1)) {
				$("#s-m-msg-block-" + self.curMsg, self.$$elmt).css("display", "none");
				self.curMsg++;
				$("#s-m-msg-block-" + self.curMsg, self.$$elmt).css("display", "block");
				self._updateButtons();
			}
		},
		_prev: function() {
			var self = this;
			if (self.curMsg > 0) {
				$("#s-m-msg-block-" + self.curMsg, self.$$elmt).css("display", "none");
				self.curMsg--;
				$("#s-m-msg-block-" + self.curMsg, self.$$elmt).css("display", "block");
				self._updateButtons();
			}
		},
		_more: function() {
			var self = this;
			var $$btn = $("#s-m-msg-stacks", self.$$elmt);
			var $$parent = $("#s-m-msg-block-" + self.curMsg);
			var $$stackSection = $("section", self.$$elmt);
			if ($$stackSection.css("display") === "none") {
				$$stackSection.css("display", "block");
				$$btn.text(locale.text("modal.diagnoses.btnstack_hide").toUpperCase());
				self._handleGesture();
			} else {
				$$stackSection.css("display", "none");
				$$btn.text(locale.text("modal.diagnoses.btnstack_show").toUpperCase());
			}
		},
		notifMainLayoutChanged: function(info, orientation, deviceType) {
			if (info.orientationChanged) {
				this._css(orientation, deviceType)
			}
		},
		_css: function(orientation, deviceType) {
			if (deviceType !== "smartphone") return;
			var $$modalDialog = $(".modal-dialog", this.$$elmt);
			if (orientation === "portrait") {
				$$modalDialog.css("margin-top", "50%")
			} else {
				$$modalDialog.css("margin-top", "")
			}
		},
		_updateButtons: function() {
			var self = this;

			for (var ii = 0, jj = self.diagId; ii < jj; ii++) {
				var $$navDot = $("a#s-m-msg-goto-" + ii + " i", self.$$elmt);
				$$navDot.toggleClass("fa-circle", ii == self.curMsg);
				$$navDot.toggleClass("fa-circle-o", ii !== self.curMsg);
			}

			// Only show x of y in case there are more than 1 diagnoses
			if (self.diagId > 1) {
				$(".modal-title", self.$$elmt).text(self.title + " (" + (self.curMsg + 1) + "/" + self.diagId + ")");
			} else {
				// hide nav circles if one diag only
				$(".s-m-modal-diag-goto", self.$$elmt).css("display", "none");
			}
		},
		_onShow: function(evt) {
			var self = this;
			self._updateButtons();
			self._css(siteLayout.getPageOrientation(), siteLayout.getDeviceType())
			self._checkSize();
			// init swype handler if more than one diagnose
			if (self.diagId > 1) {
				self._initSwiper();
			}

		},
		_handleGesture: function() {
			var self = this;
			self._gestureMgrMap = self._gestureMgrMap || {};

			if (!self._gestureMgrMap["gestureMgr" + self.curMsg]) {
				var $$scrollElmt = $(".s-m-msg-stack-content", $("#s-m-msg-block-" + self.curMsg, self.$$elmt));
				var $$wrapper = $(".s-m-scroll-wrapper", $("#s-m-msg-block-" + self.curMsg, self.$$elmt));

				self._gestureMgrMap["gestureMgr" + self.curMsg] = scroller.newScroller($$scrollElmt, {
					direction: "v"
				});
				var scrollRect = uiRect.elmtRect($$wrapper, "outer");
				self._gestureMgrMap["gestureMgr" + self.curMsg].init(scrollRect);
			}
		},
		/*
		 * User can navigate from one diagnose to another with a swipe gesture.
		 * This method will display appropriate diagnose message depending on the swipe gesture
		 */
		_onSwipe: function(evt) {
			if (hammerScroller.isEvtDirection(evt, "v")) return;

			var self = this;
			setTimeout(function() {
				var next = evt ? evt.type === "swipeleft" : null;
				if (next) {
					self._next();
				} else {
					self._prev();
				}
			});
		},
		_destroySwiper: function() {
			var self = this;
			if (self.hammer) {
				self.hammer.destroy();
				self.hammer = null;
			}
		},
		_initSwiper: function() {
			var self = this;
			self._destroySwiper();
			self.hammer = new hammerScroller.Hammer($(".s-m-modal-diag-content", self.$$elmt).get(0));
			self.hammer.get("swipe").set({
				direction: hammerScroller.Hammer.DIRECTION_HORIZONTAL
			});
			// listen to events...
			this.hammer.on("swipeleft swiperight", hammerScroller.Hammer.bindFn(self._onSwipe, self));
		},
		_checkSize: function() {
			var self = this;
			var $$app = $("#s-m-app-id");
			var $$stacks = $("message > section", self.$$elmt);
			var maxHeight = this.$$elmt.height() - (parseInt($(".modal-dialog", this.$$elmt).css("marginTop"), 10) + $(".modal-dialog", this.$$elmt).height() + 10)
			$$stacks.css({
				"max-height": maxHeight + "px"
			});
		}
	}
);
exports.Modal = _Klass;
