// syracuse-tablet/html/js/ui/modals/modalYesNoCancel.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var modalDialog = require('syracuse-tablet/html/js/ui/modals/modalDialog');
var locale = require('syracuse-tablet/html/js/helpers/locale');

var _templateAll = {
	header: '\
		{{#if title}}<div class="modal-title">{{title}}</div>{{/if}}',
	content: '\
		<div class="s-m-modal-content">\
			{{#each messages}}\
				<p>{{message}}</p>\
			{{/each}}\
		</div>',
	footer: '\
		<button type="button" class="btn btn-default" data-action="yes">{{yes}}</button> \
		<button type="button" class="btn btn-default" data-action="no">{{no}}</button> \
		<button type="button" class="btn btn-default" data-action="cancel">{{cancel}}</button>'
};

var _checkMsg = function(msg) {
	return msg ? msg.replace(/\n/g, "<br>") : "";
};


var _Klass = utils.defineClass(
	function ModalYesNoCancel(title, messages) {
		var self = this;
		self.selectedAction = "cancel";
		modalDialog.ModalBase.call(self);
		messages = messages ? typeof messages === "string" ? [messages] : messages : "";
		self.messages = [];
		(messages || []).forEach(function(m) {
			m = _checkMsg(m);
			if (m) {
				self.messages.push({
					message: m
				});
			}
		});
		self.title = title;
	}, modalDialog.ModalBase, {
		_getTemplates: function() {
			return {
				std: _templateAll
			};
		},
		getDisplayFlags: function() {
			var res = this.displayFlags || {};
			res.modalClass = "yesnocancel " + this.deviceType;
			return res;
		},
		_getDataContext: function() {
			return {
				yes: locale.text("yes"),
				no: locale.text("no"),
				cancel: locale.text("cancel"),
				title: this.title,
				messages: this.messages
			};
		},
		_onAction: function(action, param) {
			this.selectedAction = action;
			// close on actions
			return true;
		},
		_getResult: function() {
			return this.selectedAction;
		}
	}
);
exports.Modal = _Klass;
