// syracuse-tablet/html/js/ui/modals/modalShowLegal.js
"use strict";

var globals = require('syracuse-tablet/html/js/app/globals');
var utils = require('syracuse-tablet/html/js/helpers/utils');
var jsutils = require('syracuse-tablet/html/js/helpers/jsutils');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var Base = require('syracuse-tablet/html/js/ui/modals/modalDialogScroll').ModalDialogScroll;
var date = require('syracuse-tablet/html/js/helpers/types/date');
var ajax = require('syracuse-tablet/html/js/common/ajax');

var _templateAll = {
	headerContent: ' \
		<div class="s-m-header-brand"> \
			<h2 class="s-m-brand">{{label_brand}} {{label_product}}</h2> \
		</div> \
		',
	content: ' \
				<div class="form-group"> \
					<h3 class="s-m-product-update-label">{{label_update}}</h3>\
					<label class="s-m-product-update-value">{{update}}</label> \
				</div> \
				<div class="form-group"> \
					<h3>{{label_legal}}</h3> \
					<label>{{text1}}</label> \
					<label>{{text2}}</label> \
					<label>{{text3}}</label> \
					<label>{{text4}}</label> \
					<label>{{text5}}</label> \
					<label>{{text6}}</label> \
				</div> \
				{{#if licenses}} \
				<div class="form-group"> \
					<h3>{{label_license}}</h3> \
					{{#each licenses}} \
						<div class="s-m-field"><div class="s-m-title">{{../label_registrationNumber}}</div><div class="s-m-value">{{toRegNum}}</div></div> \
						<div class="s-m-field"><div class="s-m-title">{{../label_licensedTo}}</div><div class="s-m-value">{{toName}}</div></div> \
						<div class="s-m-field"><div class="s-m-title">{{../label_reference}}</div><div class="s-m-value">{{toRefNum}}</div></div> \
						<div class="s-m-field"><div class="s-m-title">{{../label_installedBy}}</div><div class="s-m-value">{{installedBy}}</div></div> \
						<div class="s-m-field"><div class="s-m-title">{{../label_expiryDate}}</div><div class="s-m-value">{{expiryDate}}</div></div> \
					{{/each}} \
				</div> \
				{{/if}} \
				{{#if userAgentBuilddate}} \
					<div class="form-group"> \
					{{#if userAgent}}\
						<h3>{{label_user_agent}}</h3> \
						<div class="form-group"> \
							<label>{{userAgent}}</label> \
						</div> \
					{{/if}} \
					{{#if builddate}} \
						<div class="form-group"> \
							<label>{{builddate}}</label> \
						</div> \
					{{/if}} \
					</div>\
				{{/if}}',
};
var _templates = {};

function _getHtml(name, ctx) {
	var tmpl = _templates[name];
	if (!tmpl) {
		tmpl = _templates[name] = Handlebars.compile(_templateAll[name]);
	}
	return tmpl(ctx);
};
var _Klass = utils.defineClass(
	function ModalShowLegal() {
		var self = this;
		Base.call(self);
	}, Base, {
		getDisplayFlags: function() {
			var opts = {
				tablet: "show-legal full",
				smartphone: "show-legal full"
			}
			return Base.prototype.getDisplayFlags.call(this, opts)
		},
		_getDataContext: function() {
			var self = this;
			return ajax.request("/sdata/syracuse/collaboration/syracuse/licenses/$service/current")
				.then(function(result) {
					var data = result.data;
					var version = null;
					var licenses = [];
					if (data && data.length > 0) {
						data.forEach(function(lic) {
							var l = {};
							l.expiryDate = locale.text("modal.legal.license.expiryDateValue", [date.getLocalShortDate(lic.validFrom), date.getLocalShortDate(lic.expiryDate)]);
							l.toName = lic.licensedTo.name;
							l.toRegNum = lic.licensedTo.registrationNumber;
							l.toRefNum = lic.licensedTo.reference;
							l.installedBy = lic.reseller.name;

							licenses.push(l);
							if (!version) {
								version = lic.productVersion;
							}
						});
						return {
							version: version,
							licenses: licenses
						};
					}
				})
				.then(function(licData) {
					var labels = locale.getProductLabels();
					var showBuildDate = window.location.href.indexOf("SHOW_VERSION_NUMBER") > -1;
					var builddate = showBuildDate && locale.text("modal.legal.builddate", [globals.getBuildTimeStamp()]);
					var userAgent = jsutils.isDvlpVersion() ? navigator.userAgent : "";
					var pageInfo;

					var data = {
						label_update: locale.text("modal.legal.update"),
						label_license: locale.text("modal.legal.about.license"),
						label_legal: locale.text("modal.legal.about.legal"),
						label_user_agent: locale.text("modal.legal.about.user_agent"),
						update: (licData && licData.version) || "?",
						licenses: licData && licData.licenses,
						builddate: builddate,
						text1: locale.text("modal.legal.text1"),
						text2: locale.text("modal.legal.text2"),
						text3: locale.text("modal.legal.text3"),
						text4: locale.text("modal.legal.text4"),
						text5: locale.text("modal.legal.text5"),
						text6: locale.text("modal.legal.text6"),
						label_registrationNumber: locale.text("modal.legal.license.registrationNumber"),
						label_licensedTo: locale.text("modal.legal.license.licensedTo"),
						label_reference: locale.text("modal.legal.license.reference"),
						label_userLicense: locale.text("modal.legal.license.userLicense"),
						label_installedBy: locale.text("modal.legal.license.installedBy"),
						label_server: locale.text("modal.legal.license.server"),
						label_expiryDate: locale.text("modal.legal.license.expiryDate"),
						userAgent: userAgent,
						userAgentBuilddate: (userAgent + builddate)
					};
					var ctx = Base.prototype._getDataContext(this);
					var content = _getHtml("content", data);
					var headerContent = _getHtml("headerContent", {
						"label_brand": labels.brand,
						"label_product": labels.product
					});
					var extendedCtx = $.extend(true, ctx, {
						title: locale.text("modal.legal.title"),
						headerContent: headerContent,
						content: content,

					})
					return extendedCtx;
				});
		},
		_onValidate: function() {},

		_onCancel: function() {},

		_getResult: function() {}
	}
);

exports.Modal = _Klass;
