// syracuse-tablet/html/js/ui/modals/modalMultiButtons.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var modalDialog = require('syracuse-tablet/html/js/ui/modals/modalDialog');

var _templateAll = {
	header: '\
		{{#if title}}<div class="modal-title">{{title}}</div>{{/if}}',
	content: '\
		<div class="s-m-modal-content">\
			<p>{{message}}</p>\
		</div>',
	footer: '\
	{{#each buttons}}\
		<button type="button" {{#if isHidden}}style="display:none"{{/if}} class="btn btn-default {{#if isDisabled}}disabled{{/if}}" data-action="{{action}}">{{label}}</button>\
	{{/each}}'
};
/**
 * Displays a dialog yes/no - yes/no/cancel ... according to opts
 * opts : {
 * 		buttons:[{action:"yes", label:"yes", isHidden: false, , isDisabled: false}],
 * 		defaultAction: "no"
 * }
 */
var _Klass = utils.defineClass(
	function ModalMultiButtons(title, message, opts) {
		modalDialog.ModalBase.call(this);
		this._dataContext = $.extend({
			title: title,
			message: message || ""
		}, opts);
		this.selectedAction = opts.defaultAction || null;
	}, modalDialog.ModalBase, {
		_getTemplates: function() {
			return {
				std: _templateAll
			};
		},
		getDisplayFlags: function() {
			var res = this.displayFlags || {};
			res.modalClass = "multibuttons " + this.deviceType;
			return res;
		},
		_getDataContext: function() {
			return this._dataContext;
		},
		_onAction: function(action, param) {
			this.selectedAction = action;
			// close on actions
			return true;
		},
		_getResult: function() {
			return this.selectedAction;
		}
	}
);
exports.Modal = _Klass;
