// syracuse-tablet/html/js/ui/modals/modalSaveDraft.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var Base = require('syracuse-tablet/html/js/ui/modals/modalMultiButtons').Modal;
var locale = require('syracuse-tablet/html/js/helpers/locale');
var siteLayout = require('syracuse-tablet/html/js/ui/siteLayout');
var modalDialog = require('syracuse-tablet/html/js/ui/modals/modalDialog');

var _template = {
	comment: '\
		<div style="display:table;">\
			<div style="display:table-row;">\
				<div style="display:table-cell;padding-right: 20px;">\
					{{comment}}\
				</div>\
				<div style="display:table-cell;width:100%;">\
					<input type="text" style="width: 100%;">\
				</div>\
			</div>\
		</div>'
};
/**
 * comment: comment input by the user
 * offline: true if user is asked to save an offline draft
 */
var _Klass = utils.defineClass(
	function ModalSaveDraft(comment, offline) {
		var opts = {
			buttons: []
		};
		["save", "cancel"].forEach(function(act) {
			opts.buttons.push({
				action: act,
				label: locale.text(act).toUpperCase()
			});
		});
		var type = offline === true ? "offline" : "online";
		Base.call(this, locale.text("drafts.dlog.save.title." + type), locale.text("drafts.dlog.save.info." + type), opts);
		this._comment = comment || "";
	}, Base, {
		_attachDOM: function() {
			Base.prototype._attachDOM.call(this);
			$(Handlebars.compile(_template.comment)({
				comment: locale.text("comment").smCapitalize()
			})).appendTo(this.$$elmt.find(".s-m-modal-content")).find('input[type="text"]').val(this._comment).focus();
		},
		_getResult: function() {
			return {
				action: Base.prototype._getResult.call(this),
				comment: this.$$elmt.find('input[type="text"]').val()
			};
		},
		_onShow: function(evt) {
			//modalDialog.ModalBase.prototype._onShow.call(this, evt);
			this._css(siteLayout.getPageOrientation(), siteLayout.getDeviceType())
		},
		notifMainLayoutChanged: function(info, orientation, deviceType) {
			if (info.orientationChanged) {
				this._css(orientation, deviceType)
			}
		},
		_css: function(orientation, deviceType) {
			if (deviceType !== "smartphone") return;
			var $$modalDialog = $(".modal-dialog", this.$$elmt);
			if (orientation === "portrait") {
				$$modalDialog.css("margin-top", "40%")
			} else {
				$$modalDialog.css("margin-top", "")
			}
		}
	}

);
exports.Modal = _Klass;
