// syracuse-tablet/html/js/ui/modals/modalSaveAuth.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var modalDialog = require('syracuse-tablet/html/js/ui/modals/modalDialog');

var _templateAll = {
	header: '\
			<buttons class="pull-right"> \
				<button type="button" class="btn btn-default s-m-modal-btn-validate" data-action="$validate"></button> \
				<button type="button" class="btn btn-default s-m-modal-btn-cancel" data-action="$cancel"></button> \
			</buttons> \
			<div class="modal-title">{{title}}</div> \
			',
	content: ' \
			<div class="s-m-scroll-wrapper s-m-scroll-wrapper-v">\
			<div class="s-m-scroll-element"> \
				<div class="form-group"> \
					<label>{{label_code}}</label> \
					<input class="form-control" id="authCode" value="{{authCodeDefault}}" /> \
				</div> \
				<div class="form-group"> \
					<label>{{label_title}}</label> \
					<input class="form-control" id="authTitle" value="{{authTitleDefault}}" /> \
				</div> \
				<div class="form-group"> \
					<label>{{label_description}}</label> \
					<input class="form-control" id="authDescription" value="{{authDescriptionDefault}}"  /> \
				</div> \
				<div class="form-group"> \
					<label>{{label_comment}}</label> \
					<input class="form-control" id="authComment" value="{{comment}}"  /> \
				</div> \
				<label>{{label_saveas_option}}</label> \
				<div class="btn-group btn-group-justified" data-toggle="buttons"> \
					<div class="btn-group"> \
						<label class="btn btn-primary active"> \
							<input type="radio" name="options" id="personal_copy" checked>{{label_personal}} \
						</label> \
					</div> \
					<div class="btn-group"> \
						<label class="btn btn-primary"> \
							<input type="radio" name="options" id="global_variant"> {{label_global}} \
						</label> \
					</div> \
					<div class="btn-group"> \
						<label class="btn btn-primary"> \
							<input type="radio" name="options" id="factory_variant">{{label_factory}} \
						</label> \
					</div> \
				</div> \
			</div> \
			</div>',
	footer: ''
};

var _Klass = utils.defineClass(
	function ModalSaveAuth(saveContext) {
		var self = this;
		modalDialog.ModalBase.call(self);

		self.saveContext = saveContext;

		// Result
		self.result = null;
	}, modalDialog.ModalBase, {
		_getTemplates: function() {
			var self = this;
			return {
				std: _templateAll
			};
		},

		getDisplayFlags: function() {
			var self = this;
			var displayFlags = self.displayFlags || {};
			if (self.deviceType === "tablet") {
				displayFlags.modalClass = "config-tile half_right";
			} else {
				displayFlags.modalClass = "config-tile full";
			}
			return displayFlags;
		},

		_getDataContext: function() {
			var self = this;
			var data = {
				title: locale.text("auth.save.title"),
				page: self.saveContext.pageName,
				authCodeDefault: self.saveContext.code,
				authTitleDefault: self.saveContext.title,
				authDescriptionDefault: self.saveContext.description,

				label_code: locale.text("auth.panel.label_code"),
				label_title: locale.text("auth.panel.label_title"),
				label_description: locale.text("auth.panel.label_description"),
				label_saveas_option: locale.text("auth.panel.label_saveas_option"),
				label_personal: locale.text("auth.panel.label_personal"),
				label_global: locale.text("auth.panel.label_global"),
				label_factory: locale.text("auth.panel.label_factory"),

				label_save: locale.text("auth.panel.label_save"),
				label_cancel: locale.text("auth.panel.label_cancel"),
				label_comment: locale.text("auth.panel.label_comment"),
				comment: self.saveContext.comment,
			};

			return data;
		},

		_onValidate: function() {
			var self = this;
			var code = self.$$elmt.find("#authCode").val();
			var title = self.$$elmt.find("#authTitle").val();
			var description = self.$$elmt.find("#authDescription").val();
			var comment = self.$$elmt.find("#authComment").val();
			self.result = {
				code: code,
				title: title,
				description: description,
				comment: comment,
				saveAs: $("input:radio:checked", self.$$elmt)[0].id
			};
		},

		_onCancel: function() {
			var self = this;
			self.result = null;
		},

		_getResult: function() {
			var self = this;
			return self.result;
		}
	}
);

exports.Modal = _Klass;
