// syracuse-tablet/html/js/ui/modals/modalDialogScroll.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var modalDialog = require('syracuse-tablet/html/js/ui/modals/modalDialog');

var _templateAll = {
	header: '\
		<buttons class="pull-left"> \
			<button type="button" class="btn btn-default fa fa-arrow-left" data-action="$cancel"></button> \
		</buttons> \
		{{#if buttonValidate }}\
			<buttons class="pull-right"> \
				<button type="button" class="btn btn-default fa fa-check " data-action="$validate"></button> \
			</buttons> \
		{{/if}}\
		<div class="modal-title">{{title}}</div> \
		',
	headerContent: '{{{headerContent}}}',
	content: ' \
			<div class="s-m-scroll-wrapper s-m-scroll-wrapper-v">\
				<div class="form-group s-m-scroll-element"> \
					{{{content}}}\
				</div> \
			</div>',
	footer: '{{{footer}}}'
};

exports.ModalDialogScroll = utils.defineClass(
	function ModalDialogScroll(appList) {
		var self = this;
		modalDialog.ModalBase.call(self);
	}, modalDialog.ModalBase, {
		_getTemplates: function() {
			return {
				std: _templateAll
			};
		},
		getDisplayFlags: function(opts) {
			var self = this;
			var displayFlags = self.displayFlags || {};
			displayFlags.modalClass = opts[self.deviceType];
			return displayFlags;
		},

		_getDataContext: function() {
			return {
				label_validate: locale.text("welcome.select.app.add"),
				label_cancel: locale.text("welcome.select.app.cancel"),
			};
		}
	}
);
