// syracuse-tablet/html/js/ui/modals/modal.js
"use strict";

var uiutils = require('syracuse-tablet/html/js/ui/uiUtils');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var globals = require('syracuse-tablet/html/js/helpers/globals');
var jsutils = require('syracuse-tablet/html/js/helpers/jsutils');
var native = require('syracuse-tablet/html/js/helpers/native/native');

// Separated module for showing diagnoses since it may be replaced depending on device/OS
var ModalDiagnoses = require('syracuse-tablet/html/js/ui/modals/modalDiagnoses').Modal;
var ModalConfirm = require('syracuse-tablet/html/js/ui/modals/modalConfirm').Modal;
var ModalField = require('syracuse-tablet/html/js/ui/modals/modalField').Modal;

var _checkMsg = function(msg) {
	return msg ? msg.replace(/\n/g, "<br>") : "";
};
var _templates = {
	confirm: ' \
		<div class="modal fade" id="modal-confirm" tabindex="-1" role="dialog"> \
			<div class="modal-dialog s-m-modal-layout">\
				<header>\
					<button type="button" class="close" data-dismiss="modal" aria-hidden="true">&#215;</button> \
				</header>\
				<section>\
					<div class="s-m-modal-content row">\
						<div class="col-md-4"></div>\
						<div class="col-md-4">\
							<header>{{title}}</header>\
							<section>{{message}}</section>\
							<footer>\
								<button type="button" class="btn btn-default" data-action="yes">{{yes}}</button> \
				                <button type="button" class="btn btn-default" data-action="no">{{no}}</button> \
							</footer>\
						</div>\
					<div class="col-md-4"></div>\
				</div>\
				</section>\
			</div> \
		</div>',
	error: ' \
		<div class="modal fade" id="modal-error" tabindex="-1" role="dialog"> \
			<div class="modal-dialog s-m-modal-layout">\
				<header>\
					<button type="button" class="close" data-dismiss="modal" aria-hidden="true">&#215;</button> \
				</header>\
				<section>\
					<div class="s-m-modal-content row">\
						<div class="col-md-4"></div>\
						<div class="col-md-4">\
							<header class="error">{{title}}</header>\
							<section>{{body}}</section>\
							<footer><button type="button" class="btn btn-default" data-dismiss="modal">{{close}}</button></footer>\
						</div>\
						<div class="col-md-4"></div>\
					</div>\
				</section>\
			</div> \
		</div>',
	info: '\
		<div class="modal fade" id="modal-info" tabindex="-1" role="dialog"> \
		    <div class="modal-dialog s-m-modal-layout">\
		    	<header>\
		    		<button type="button" class="close" data-dismiss="modal" aria-hidden="true">&#215;</button> \
		    	</header>\
		    	<section>\
		    		<div class="s-m-modal-content row">\
			    		<div class="col-md-4"></div>\
			    		<div class="col-md-4">\
			    			<header class="info">{{title}}</header>\
			    			<section>{{body}}</section>\
			    			<footer><button type="button" class="btn btn-default" data-dismiss="modal">{{close}}</button></footer>\
			    		</div>\
			    		<div class="col-md-4"></div>\
		    		</div>\
		    	</section>\
			</div> \
		</div>',
	field: '\
		<div class="modal fade" id="modal-field" tabindex="-1" role="dialog"> \
		    <div class="modal-dialog"> \
				<div class="modal-content"> \
				    <div class="modal-header"> \
						<button type="button" class="close" data-dismiss="modal" aria-hidden="true">&#215;</button> \
						<h4 class="modal-title" id="myModalLabel">{{title}}</h4> \
				    </div> \
				    <div class="modal-body"> \
				       {{{body}}} \
				    </div> \
				    <div class="modal-footer"> \
				        <button type="button" class="btn btn-default" data-dismiss="modal">{{close}}</button> \
					</div> \
				</div> \
			</div> \
		</div>',
	actionMessage: '\
		<div class="s-m-act-msg {{severityClass}}">\
			<a draggable="false" href="#"><span class="glyphicon glyphicon-remove"></span></a>\
			<header>\
				<div>{{title}}</div>\
			</header>\
			<section>\
				<article>{{message}}</article>\
			</section>\
			<footer></footer>\
		</div>'
};

var _getHtml = function(name, ctx) {
	var key = name + "compiled";
	var tmpl = _templates[key];
	if (!tmpl) {
		tmpl = _templates[key] = Handlebars.compile(_templates[name]);
	}
	return tmpl(ctx);
};

/* First param : cb - Second msg - Third data ...*/
var _execCb = function(cb, msg, data) {
	if (!cb || arguments.length == 0) return;
	var cb = arguments[0];
	var args = Array.prototype.slice.call(arguments, 1);
	cb.apply(null, args);
};


/************************************************************************
 ** IMPORTANT
 * The code of the callback parameter in the exported methods :
 * error(), info(), diagnoses(), field(), confirm(), simpleconfirm(), modal()
 * has to indicate whether the callback is called :
 * - when the modal opens
 * - when the modal closes
 * - at both events
 *
 * Here is an example :
 * function (modalEvent, $$modalElmt) {
 *        // build content on open
 *       if (modalEvent == 'shown.bs.modal') {
 *           // code which is run on modal opening
 *       }
 *
 *       // run actions on close 'hidden.bs.modal'
 *       else {
 *       	// code which is run on modal closing
 *       }
 * });
 *************************************************************************/



/*
 * Examples
 * 
 * error("There was an error");
 * 
 * error("There was an error", "I am the detail text");
 *
 * error("There was an error", new Error("I am an exception"));
 * 
 * error("There was an error", {
 *   $diagnoses: [{
 *	   $message: "Message 1",
 *     $severity: "error",
 *     $stackTrace: "bla"
 *	 }, {
 * 	   $message: "Message 2",
 *	   $severity: "info",
 * 	   $stackTrace: "bla"
 *   }]
 * });
 * 
 * error("There was an error", {
 *   message: "I am the message", 
 *   detail: "I am the detail message"
 * });

 * error("There was an error", {
 *   message: "I am the message", 
 *   detail: new Error("I am an exception")
 * });
 * 
 */
var _error = function(title, e, cb) {
	var diag;
	diag = jsutils.convertToDiagnoses(title, e);
	if (title == null && e && e.$title) {
		title = e.$title;
	}
	return _diagnoses(title, diag, cb);
};

/*
 * title. Title of dialog
 * diag: Object containing diagnoses
 *       {
 *           $diagnoses: [{
 *               $message: "bla...",
 *               $severity: "error", // success, info, warning, error
 *               $stackTrace: "Stacktrace as string, formatted"
 *           }, {
 *               ...
 *           }]
 *       }
 * cb: Callback
 */
var _diagnoses = function(title, diag, cb) {
	var modal = new ModalDiagnoses(title, diag);
	modal.show().then(function(result) {
		if (cb) {
			cb(result);
		}
	});
	//return modalDiag.show(title, diag, cb);
};

/* 
 * Wrap info text into diagnoses and display it in the generic diagnoses dialog
 */
var _info = function(title, text, cb) {
	var infoDiag = {
		"$diagnoses": [{
			"$severity": "info",
			"$message": text
		}]
	};
	_diagnoses(title, infoDiag, cb);
};

var _confirm = function(action, cb) {

	var modal = new ModalConfirm(null, action);

	modal.show().then(function(result) {
		if (cb && (result === "yes" || result === "no")) {
			cb(result, "yes" === result);
		}
	});
};

var _simpleConfirm = function(message, cb) {
	var modal = new ModalConfirm(message, null, true);

	modal.show().then(function(result) {
		if (cb && (result === "yes" || result === "no")) {
			cb(result, "yes" === result);
		}
	});
};

/*
 * simpleConfirmPromise({
 *   message: "Are you sure?"
 *   }).then(function(result) {
 *      if (result.action === "yes") {
 *      } else if (result.action === "no") {
 *      }
 *   });
 */
exports.simpleConfirmPromise = function(options) {
	var deferred = $.Deferred();
	var modal = new ModalConfirm(options.message, null, true);

	modal.show().then(function(result) {
		if (result === "yes" || result === "no") {
			deferred.resolve({
				action: result
			});
		}
	});

	return deferred.promise();
};
/*
 * Open a modal window linked to a field
 */
var _field = function(title, domBody, cb) {
	var modal = new ModalField(title, domBody, cb);
	modal.show();
};

var _actionResult = function(title, obj, forceToast) {
	obj = jsutils.convertToDiagnoses(obj);
	if (!obj || !obj.$diagnoses || obj.$diagnoses.length == 0) return;
	var hasStack = false,
		hasError = false;
	var text = [];
	obj.$diagnoses.forEach(function(diag) {
		text.push(diag.$message);
		if (diag.$severity == "error") {
			hasError = true;
			hasStack = hasStack || diag.$stackTrace != null;
		}
	});
	if (!hasStack || forceToast === true) {
		var options = {
			title: title,
			body: text.join('\n'),
			severityClass: hasError ? "error" : "success",
			$$parentSlot: globals.getApplication().getMessageSlot()
		};

		if (!options.onlyWebapp && native.hasCapability("toastMessages")) {
			// Adds the message on the toast message stack on windows
			native.getModule("toastMessages").showToastMessage({
				Text1: options.title,
				Text2: options.body,
				Severity: hasError ? "error" : "success"
			});
		}

		_notificationHandler(options);
	} else {
		_diagnoses(title, obj);
	}
};

/* options properties are :
- severityClass : css class depending on severity
- title : title text value used in the template
- text : body text value used in the template
- $$parentSlot : parent to which modal is appended, depends on device type
*/
var _notificationHandler = function(options) {

	// create html
	var h = _getHtml("actionMessage", {
		title: options.title,
		message: options.body,
		severityClass: options.severityClass
	});

	var isClosing;
	// append to app according to device type
	var $$actionModal = $(h);
	options.$$parentSlot.prepend($$actionModal);

	// show parentSlot
	options.$$parentSlot.show();

	// close anchor
	var $$closer = $$actionModal.find("a");

	function toggle() {
		$$actionModal.toggleClass("show");
	}

	function clear() {
		toggle();
		// set timeout in order to have css animation visible before removing DOM
		setTimeout(function() {
			$$actionModal.remove();

			// hide parent slot if last item removed
			if (options.$$parentSlot.children().length == 0) {
				options.$$parentSlot.hide();
			}
		}, 500);
	};

	// bind click action on closer button. triggers close action
	$$closer.bind("click", function() {
		clear();
		isClosing = true;
	});

	// display
	// setTimeout is necessary to have the display css animation working
	setTimeout(function() {
		toggle();
	}, 50);

	// toggle after 5 seconds
	var toggleTimeout = setTimeout(function() {
		if (!isClosing) {
			clear();
		}
	}, 5000);
};


var _modal = function(html, cb) {
	// Disable wait wheel if any
	uiutils.waitWheelStop();
	var app = globals.getApplication();
	if (app && app.currentPage) {
		app.currentPage.waitWheelStop();
	}
	/* Always to retry read template and display alert if not found */
	var h = $(html).modal({
		keyboard: false,
		backdrop: "static"
	});
	var viewportHeight = $(window).height();
	$('.modal-body', h).css('max-height', viewportHeight * 0.7);

	// move modal within the app dom element
	var $$md = $(h);
	var $$parent = $("#s-m-modal-parent") && $("#s-m-modal-parent").length > 0 ? $("#s-m-modal-parent") : $("<div>").attr("id", "s-m-modal-parent");
	$$parent.append($$md);
	$("#s-m-app-id").append($$parent);
	$$parent.toggle(true);

	var done = false;


	// Use <button data-dismiss"modal" for close action
	// Use <button data-action"myaction" for action other than
	// close
	h.on("shown.bs.modal", function(evt) {
		try {
			_execCb(cb, "shown.bs.modal", h);
		} catch (e) {
			alert("Modal open - call back error" + e.stack);
		}
		// !!We need a synchronous call
		app.notifModalOpen();
	});
	// hide.bs.modal is better than hidden.bs.modal be cause it comes sooner
	// Like open modals are accounted in aplication it's better to notifModalClose ASAP otherwise changePage doesn't work fine (no changePage if a modal is opened)
	app.notifModalClose();
	h.on("hide.bs.modal", function() {
		$("#s-m-modal-parent").toggle(false);

		// !!We need a synchronous call
		app.notifModalClose();
		// Call cb only if no click action
		if (!done && cb) {
			try {
				_execCb(cb, "hide.bs.modal", h);
			} catch (e) {
				alert("Modal close - call back error" + e.stack);
			}
		}
		// Called when dialog is close - data-dismiss = "modal"
		// or m.modal('hide')
		h.unbind();
		h.remove();
	});
	// For dialogs with option
	h.delegate("button[data-action]", "click", function(evt) {
		var t = $(evt.target);
		var act = t.attr("data-action");
		if (act) {
			done = true;
			evt.preventDefault();
			evt.stopPropagation();
			_execCb(cb, act, h, t.attr("data-params"));
		}
		// Close modal
		h.modal('hide');
	});
	return h;
};

/**
 * options.onlyWebapp = true if set, notification is not send to operation system but only showen inside application
 */
exports.notify = function(options) {
	options.$$parentSlot = options.$$parentSlot || (globals.getApplication() && globals.getApplication().getMessageSlot());
	_notificationHandler(options);

	if (!options.onlyWebapp && native.hasCapability("toastMessages")) {
		// Adds the message on the toast message stack on windows
		native.getModule("toastMessages").showToastMessage({
			Text1: options.title,
			Text2: options.body,
			Severity: options.severityClass
		});
	}
};

var _displayActionMsg = function(success, name, proto, data) {
	var obj;
	if (data && data.$diagnoses && data.$diagnoses.length > 0) {
		obj = data;
	} else {
		obj = {
			$diagnoses: [{
				$message: locale.text("modal.action." + (success ? "success" : "fail")),
				$severity: success ? "success" : "error",
				$stackTrace: JSON.stringify(data)
			}]
		};
	}
	var title = name;
	if (proto) {
		var links = proto.data("$links");
		var link = links && links[name];
		if (link && link.$title) {
			title = proto.resolveExpression(link.$title);
		}
	}
	_actionResult(title, obj);
};

exports.error = _error;
exports.info = _info;
exports.diagnoses = _diagnoses;
exports.field = _field;
exports.confirm = _confirm;
exports.simpleConfirm = _simpleConfirm;
exports.modal = _modal;
exports.actionResult = _actionResult;
exports.displayActionMsg = _displayActionMsg;
