// syracuse-tablet/html/js/ui/modals/modalConfirm.js
"use strict";

var globals = require('syracuse-tablet/html/js/helpers/globals');
var utils = require('syracuse-tablet/html/js/helpers/utils');
var modalDialog = require('syracuse-tablet/html/js/ui/modals/modalDialog');

var _templateAll = {
	header: '\
		<div class="modal-title">{{title}}</div>',
	content: '\
			<div class="s-m-modal-content">\
				{{message}}
			</div>',
	footer: '\
		<button type="button" class="btn btn-default" data-action="yes">{{yes}}</button> \
		<button type="button" class="btn btn-default" data-action="no">{{no}}</button>'
};

var _Klass = utils.defineClass(
	function ModalDiagnoses(title, diag, cb) {
		var self = this;
		modalDialog.ModalBase.call(self);
		// Init params
		self.title = title;
		self.diag = diag;
		self.cb = cb;
	}, modalDialog.ModalBase, {
		_getTemplates: function() {
			var self = this;
			return {
				std: _templateAll
			};
		},
		getDisplayFlags: function() {
			var self = this;
			var displayFlags = self.displayFlags || {};
			if (self.deviceType === "tablet") {
				displayFlags.modalClass = "confirm tablet";
			} else {
				displayFlags.modalClass = "confirm mobile";
			}
			return displayFlags;
		},
		_getDataContext: function() {
			var self = this;
			var ctx = {};
		}
	}
);
exports.Modal = _Klass;
