// syracuse-ui/lib/common/controller/localize.js
"use strict";
exports.site = {
    logout: "Logout",
    login: "Login"
};
exports.field = {
    choiceFieldNoMatch: "no match"
};
exports.datePicker = {
    today: "Today",
    week: "Week",
    quickMenu: {
        begin: "Begin",
        end: "End",
        interval: {
            ofYear: "of year",
            ofQuarter: "of quarter",
            ofMonth: "of month",
            ofWeek: "of week"
        }
    },
    ok: "Ok",
    cancel: "Cancel"
};
exports.timePicker = {
    title: "Time Picker"
};

exports.list = {
    select: "Select",
    noRecords: "No records to display",
    link: "Add",
    filter: {
        clear: "Clear",
        and: "and"
    },
    pager: {
        totalRecords: "record(s)",
        itemOf: "of",
        first: "First",
        last: "Last",
        prev: "Prev",
        next: "Next"
    }
};
exports.diagnoseManager = {
    error: "Error(s)",
    fatal: "Fatal(s)",
    warning: "Warning(s)",
    info: "Information(s)",
    showAll: "Show All",
    collapse: "Collapse"
};
exports.messageBox = {
    info: "Information",
    confirm: "Confirmation",
    warning: "Warning",
    alert: "alert"
};
