// syracuse-ui/lib/dialog/dialog.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var WidgetResizer = require("syracuse-ui/lib/site/resize/widgetResizer").WidgetResizer;

function _resizeOverlay(dlg) {
	var rect = syra_site.getLayoutSize();
	if (dlg._content.isFusionPage) {
		var size = syra_site.getBodySize();
		dlg.overlay.style.top = size.top + "px";
		dlg.overlay.style.height = size.height + "px";
	} else {
		dlg.overlay.style.height = rect.height + "px";
	}
	dlg.overlay.style.width = rect.width + "px";
	if (!dlg.isDeactivated) {
		if (dlg._content.layoutSlot.style.display !== "none") // manage of multiple window under modal
			dlg.overlay.style.display = "";
	}
}

function _resizePositioned(dlg, siteBodyRect) {
	var slot = dlg.dialogSlot;
	var rect = syra_site.dom.getBoundingClientRect(slot);
	if (rect.width > siteBodyRect.width) {
		slot.style.width = (rect.width = siteBodyRect.width) + "px";
	}
	if (rect.height > siteBodyRect.height) {
		slot.style.height = (rect.height = siteBodyRect.height) + "px";
	}
	var $position = {
		my: dlg.options.position.my,
		at: dlg.options.position.at,
		of: dlg.options.position.of,
	};
	var page = dlg.openerBox && dlg.openerBox.page;
	if (page && page.scrollview) {
		var isWithIn = true;
		if (dlg.options.position.of) {
			isWithIn = page.scrollview.contains(dlg.options.position.of[0]);
		}
		if (isWithIn) {
			$position.collision = "flipfit";
			$position.within = $(page.scrollview);
		}
	}
	$(slot).position($position);
	dlg.options.onresize && dlg.options.onresize(dlg);
}

function _setResizer(dlg) {
	dlg._widgetResizer = new WidgetResizer();
	dlg._widgetResizer.setResizable({
		source: dlg,
		slot: dlg.dialogSlot,
		dragSpot: dlg.options.dragSpot,
		isResizingEnabled: dlg.options.isResizingEnabled,
		onResize: function(resizer, moving) {
			if (moving) {
				dlg.resizeDialog(true, null, true);
			}
		}
	});
}

function _setAutoFocus(dlg) {
	var link = document.createElement("a");
	link.className = "s-dialog-focus-anchor";
	link.setAttribute('href', '#');
	dlg.dialogSlot.appendChild(link);
	link.focus();
}

function _addOverlay(dlg) {
	if (dlg.isModal) {
		dlg.overlay = document.createElement("div");
		dlg.overlay.className = "s-overlay";
		syra_site.dom.setZIndex(dlg.overlay);
		syra_site.layoutSlot.appendChild(dlg.overlay);
	}
}

function _openContent(dlg) {
	dlg._content = dlg.options.content;
	dlg.dialogSlot = dlg.options.slot || dlg.options.content.domItem;
	if (dlg.dialogSlot.style.position != "fixed") {
		dlg.dialogSlot.style.position = "absolute";
	}
	dlg.id = dlg._content.id + "-dlg";
	syra_store.add(dlg);
	dlg.dialogSlot.style.display = '';
	dlg.options.$isAutoClose = dlg.options.$isAutoClose !== false;
	_endOpen(dlg);
	!dlg._content.isPageLoaded && dlg.resizeDialog(); //content is not page => resize
}

function _loadPageRepresentation(dlg) {
	if (dlg.options.$method && (dlg.options.$method != "GET"))
		syra_controller.postQuery(dlg.options, null, dlg.options.article, function($location, data) {
			dlg.hasCallback = true; // in order not to dispose the dialog before the callback
			syra_controller.loadRepresentation({
				article: dlg.options.article,
				segments: $location.$url,
				success: function($itemPage) {
					if (data && syra_controller.isWorkingCopyUrl(dlg.options.$url)) {
						$itemPage.initData = data;
					}
					_openPage(dlg, $itemPage);
					dlg.hasCallback = null;
				}

			});
		});
	else {
		dlg.hasCallback = true; // in order not to dispose the dialog before the callback
		syra_controller.loadRepresentation({
			article: dlg.options.article,
			segments: dlg.options.$url,
			success: function($itemPage) {
				_openPage(dlg, $itemPage);
				dlg.hasCallback = null;
			}
		});
	}
}

function _openPage(dlg, $itemPage) {
	if ($itemPage || dlg.options.page) {
		if (dlg.options.page) {
			dlg._content = dlg.options.page;
			dlg.dialogSlot = dlg.options.page.layoutSlot || document.createElement("div");
			dlg._content.dialogWrapper = dlg;
		} else {
			dlg.dialogSlot = $itemPage.layoutSlot = document.createElement("div");
			$itemPage.dialogWrapper = dlg;
			dlg._content = syra_site.pageLoader.load($itemPage);
		}
		dlg.id = dlg._content.id + "-dlg";
		syra_store.add(dlg);
		dlg.dialogSlot.className = dlg.$skin + "-page-container";
		dlg._content.header.className = dlg.$skin + "-page-head";
		dlg._content.body.className = dlg.$skin + "-page-body";
		dlg._content.dataSlot.className = dlg.$skin + "-page-data";
		dlg._content.domItem.className = dlg.$skin + "-page";
		dlg._content.domTitle.className = dlg.$skin + "-page-title";
		dlg.dialogSlot.setAttribute("tabindex", "2");
		dlg.options.$isAutoClose = dlg.options.$isAutoClose === true;
		dlg.appendCloseButton(dlg._content.header);

		if (dlg.options.onValidate) {
			dlg.btn_ok = syra_menus.addIconButton(syra_local.dlg_ok, dlg.$skin + "-page-ok", "onOkClick");
			dlg.btn_ok.syraTool = dlg.id;
			dlg._content.header.appendChild(dlg.btn_ok);
			if (dlg.options.$isOkHidden) {
				dlg.hideOkButton(true);
			}
		}
		if (dlg.options.onSelectRecord) {
			dlg._content.onSelectRecord = function(selectedRecords) {
				if (dlg.options.onSelectRecord(selectedRecords, dlg._content) !== false) {
					dlg.close();
				}
			};
		}
		dlg.options.dragSpot = dlg._content.header;
		_endOpen(dlg);
		dlg.resizeDialog();
		dlg.isDialogPage = true;
	} else {
		if (dlg.options.$itemPage) {
			_openPage(dlg, dlg.options.$itemPage);
		} else {
			_loadPageRepresentation(dlg);
		}
	}
}

function _endOpen(dlg) {
	dlg.isModal && _addOverlay(dlg);
	syra_site.dom.setZIndex(dlg.dialogSlot);
	syra_site.layoutSlot.appendChild(dlg.dialogSlot);
	dlg.dialogSlot.style.top = "0px";
	dlg.options.dragSpot && _setResizer(dlg);
	dlg.isOpened = true;
	dlg.options.$autoFocus && _setAutoFocus(dlg);
	dlg.options.onOpened && dlg.options.onOpened(dlg._content);
}


function _resizePage(dlg, page, isDlgResizing) {
	var scrollTop = page.scrollview.scrollTop;
	var maxPageHeight = page.scrollview.scrollHeight;
	var maxPageWidth = Math.max(page.scrollview.scrollWidth, page.header.scrollWidth);
	var barWidth = 0;
	if (page.menuBar) {
		maxPageHeight = Math.max(page.menuBar._bar.scrollHeight, maxPageHeight);
		barWidth += page.menuBar._bar.clientWidth;
	}
	if (page.fusionBar) {
		maxPageHeight = Math.max(page.fusionBar._bar.scrollHeight, maxPageHeight);
		barWidth += page.fusionBar.clientWidth;
	}
	if (page.designer) {
		if (page.designer.treesBar) {
			maxPageHeight = Math.max(page.designer.treesBar._bar.scrollHeight, maxPageHeight);
			barWidth += page.designer.treesBar._bar.clientWidth;
		}
		maxPageHeight = Math.max(page.designer.propertiesBodySlot.scrollHeight, maxPageHeight);
		barWidth += page.designer.propertiesBodySlot.clientWidth;
	}

	var maxDlgHeight = Math.max(100, dlg._maxHeight);
	var maxDlgWidth = Math.max(200, dlg._maxWidth);
	page.barHeight = Math.min(maxPageHeight, maxDlgHeight);
	page.scrollview.style.height = (page.barHeight - dlg._content.diagnoseSlot.getBoundingClientRect().height) + "px";
	if (page.scrollview.scrollHeight - page.scrollview.clientHeight) {
		maxPageWidth += 17;
	}
	page.scrollview.style.width = Math.min(maxPageWidth, maxDlgWidth - barWidth) + "px";

	!isDlgResizing && dlg.centerDialog();
	delete page.scrollviewSize;
	page.designer && page.designer.resizeArticle();
	page.menuBar && page.menuBar.resizeBar();
	page.fusionBar && page.fusionBar.resizeBar();
	if (isDlgResizing) {
		page.resizeArticle();
	}
	var updated;
	var savHeight = page.scrollview.clientHeight;
	page.scrollview.style.height = "";
	var newHeight = page.scrollview.clientHeight;
	if ((savHeight - newHeight) > 2) { //margin
		page.scrollview.style.height = newHeight + "px";
		updated = true;
	} else {
		page.scrollview.style.height = savHeight + "px";
	}
	var savWidth = page.scrollview.clientWidth;
	page.scrollview.style.width = "";
	var newWidth = page.scrollview.clientWidth;
	if ((savWidth - newWidth) > 2) { //margin
		page.scrollview.style.width = newWidth + "px";
		updated = true;
	} else {
		page.scrollview.style.width = savWidth + "px";
	}
	if (updated) {
		dlg.centerDialog();
	}
	if (scrollTop) {
		page.scrollview.scrollTop = Math.min(scrollTop, page.scrollview.scrollHeight - page.scrollview.clientHeight);
	}
}


function _resizeLookupPage(dlg, page, isDlgResizing) {
	var lookupField = page.lookupField;
	if (!isDlgResizing) {
		lookupField.bodyTableSlot.style.width = "";
		lookupField.bodyTableSlot.style.height = "";
	}
	var scrollTop = page.scrollview.scrollTop;
	var maxPageHeight = page.scrollview.scrollHeight;
	var maxPageWidth;
	if (lookupField.colsWidth) {
		var offset1 = page.getScrollviewSize();
		var offset2 = page.lookupField.bodyTableSlot.getBoundingClientRect();
		maxPageWidth = page.lookupField.colsWidth + 17 + (offset2.left - offset1.left) + (offset1.right - offset2.right);
	} else {
		maxPageWidth = page.scrollview.scrollWidth; // + 2; //margin
	}
	maxPageWidth = Math.max(maxPageWidth, page.header.scrollWidth);
	var barWidth = 0;
	if (page.menuBar) {
		maxPageHeight = Math.max(page.menuBar._bar.scrollHeight, maxPageHeight);
		barWidth += page.menuBar._bar.clientWidth;
	}
	if (page.designer) {
		if (page.designer.treesBar) {
			maxPageHeight = Math.max(page.designer.treesBar._bar.scrollHeight, maxPageHeight);
			barWidth += page.designer.treesBar._bar.clientWidth;
		}
		maxPageHeight = Math.max(page.designer.propertiesBodySlot.scrollHeight, maxPageHeight);
		barWidth += page.designer.propertiesBodySlot.clientWidth;
	}

	var maxDlgHeight = Math.max(100, dlg._maxHeight);
	var maxDlgWidth = Math.max(200, dlg._maxWidth);
	page.barHeight = Math.min(maxPageHeight, maxDlgHeight);
	page.scrollview.style.height = (page.barHeight - dlg._content.diagnoseSlot.getBoundingClientRect().height) + "px";
	if (page.scrollview.scrollHeight - page.scrollview.clientHeight) {
		maxPageWidth += 17;
	}
	page.scrollview.style.width = Math.min(maxPageWidth, maxDlgWidth - barWidth) + "px";

	!isDlgResizing && dlg.centerDialog();
	delete page.scrollviewSize;
	page.designer && page.designer.resizeArticle();

	if (isDlgResizing) {
		page.resizeArticle();
	}

	var diff = page.scrollview.scrollHeight - page.scrollview.clientHeight;
	if (diff) {
		diff += 2; // margin for chrome
		lookupField.bodyTableSlot.style.height = (lookupField.bodyTableSlot.getBoundingClientRect().height - diff) + "px";
	}
	diff = page.scrollview.scrollWidth - page.scrollview.clientWidth;
	if (diff) {
		diff += 2; // margin for chrome
		lookupField.bodyTableSlot.style.width = (lookupField.bodyTableSlot.getBoundingClientRect().width - diff) + "px";
	}
	lookupField.resizeScrollView && lookupField.resizeScrollView();


	var updated;
	var savHeight = page.scrollview.clientHeight;
	page.scrollview.style.height = "";
	var newHeight = page.scrollview.clientHeight;
	if ((savHeight - newHeight) > 2) { //margin
		page.scrollview.style.height = newHeight + "px";
		updated = true;
	} else {
		page.scrollview.style.height = savHeight + "px";
	}
	var savWidth = page.scrollview.clientWidth;
	page.scrollview.style.width = "";
	var newWidth = page.scrollview.clientWidth;
	if ((savWidth - newWidth) > 2) { //margin
		page.scrollview.style.width = newWidth + "px";
		updated = true;
	} else {
		page.scrollview.style.width = savWidth + "px";
	}
	if (updated) {
		dlg.centerDialog();
	}
	if (scrollTop) {
		page.scrollview.scrollTop = Math.min(scrollTop, page.scrollview.scrollHeight - page.scrollview.clientHeight);
	}

}


function Dialog() {}

exports.Dialog = helpers.defineClass(Dialog, null, {
	centerDialog: function() {
		var size = syra_site.getLayoutSize();
		var bodyH = Math.max(size.height - 50, 100);
		var bodyW = Math.max(size.width - 50, 100);
		var slot = this.dialogSlot;
		slot.style.height = slot.style.width = "";
		var slotRect = slot.getBoundingClientRect();
		var height = Math.ceil(Math.min(slotRect.height, bodyH));
		var width = Math.ceil(Math.min(slotRect.width, bodyW));
		slot.style.width = width + "px";
		slot.style.height = height + "px";
		if (!this.isUserPosition) {
			slot.style.top = ((size.height - height) / 2) + "px";
			slot.style.left = ((size.width - width) / 2) + "px";
		}
	},
	open: function(options) {
		if ((this.options = options).content) {
			_openContent(this);
		} else {
			_openPage(this);
		}
	},
	close: function(validated, dispose) {
		if (!this.disposed && this.options && this.options.onClose ? (this.options.onClose(validated !== true, dispose) !== false) : true) {
			this.dispose();
		}
	},
	hideOkButton: function(hide) {
		if (this.btn_ok) {
			this.btn_ok.style.display = hide ? "none" : "";
		}
	},
	onOkClick: function() {
		if (this.options.onValidate(this._content) !== false) {
			this.close(true);
		}
	},
	onCloseClick: function() {
		this.close();
	},
	appendCloseButton: function(slot) {
		this.btn_close = syra_menus.addIconButton(syra_local.dlg_close, this.$skin + "-page-close", "onCloseClick", "ESC");
		this.btn_close.syraTool = this.id;
		slot.appendChild(this.btn_close);
	},
	activate: function(isActivated) {
		this.isDeactivated = !isActivated;
		if (this.overlay) {
			this.overlay.style.display = isActivated ? "" : "none";
		}
		this.dialogSlot.style.display = isActivated ? "" : "none";
	},
	resizeDialog: function(resize, onScroll, isDlgResizing) {
		if (this._content) {
			if (!onScroll && this.overlay) {
				_resizeOverlay(this);
			}
			if (!this.disposed && syra_site.body) {
				var size = syra_site.getLayoutSize();
				var rect = {
					left: 50
				};
				rect.width = size.width - (rect.left * 2);
				rect.right = rect.left + rect.width;
				rect.top = 50;
				this._maxHeight = rect.height = size.height - (rect.top * 2);
				rect.bottom = rect.top + rect.height;
				this._maxWidth = rect.width;
				if (this.options.resizeDialog) {
					this.options.resizeDialog(this, isDlgResizing);
				} else {
					if (this._widgetResizer) {
						this._widgetResizer.maxHeight = this._maxHeight;
					}
					if (this.options.position) {
						_resizePositioned(this, rect);
					} else {
						var page = this._content && this._content.page;
						if (page && page.scrollview) {
							page.lookupField ? _resizeLookupPage(this, page, isDlgResizing) : _resizePage(this, page, isDlgResizing);
						}

					}
				}
			}
		}
	},
	dispose: function() {
		syra_store.remove(this);
		this.overlay && syra_site.dom.removeChild(this.overlay);
		if (this._widgetResizer) {
			this._widgetResizer.dispose();
			this._widgetResizer = null;
		}
		if (this._content) {
			delete this._content.dialogWrapper;
		}
		syra_site.dialogManager.close(this);
		if (this.options) {
			this.options.dragSpot = this.options.page = this.options.autoCloseBoundary = null;
			if (this.options.content) {
				delete this.options.content;
			} else {
				this._content.dispose && this._content.dispose();
				this.dialogSlot && syra_site.dom.empty(this.dialogSlot);
			}
			this.dialogSlot && syra_site.dom.removeChild(this.dialogSlot);
			this.options.article = this.options.onValidate = this.options.onSelectRecord = this.options.onresize = null;
		}
		this.btn_close = this.btn_ok = this.dialogSlot = this.overlay = this._content = this.options = null;
	}
});
