// syracuse-ui/lib/dialog/tip.js
"use strict";

function _show(tip, text, target) {
	if (!tip._item) {
		tip._item = syra_dom.addDiv(tip.$css);
	}
	if (text.indexOf("<em") >= 0) {
		if (!tip.em) {
			tip.em = "<em class='" + tip.$css + "-em'>";
		}
		tip._item.innerHTML = text.replace(/<em>/g, tip.em);
	} else {
		tip._item.textContent = text;
	}
	tip._item.style.zIndex = syra_dom.getTopZIndex();
	if (target.syraShortCutTip) {
		syra_dom.addTextInfo("s-title-shortcut", "(" + target.syraShortCutTip + ")", tip._item);
	}
	var posX = syra_site.isRTL ? "right" : "left";
	syra_site.layoutSlot.appendChild(tip._item);
	$(tip._item).position({
		my: posX + " top",
		at: posX + " bottom+15",
		of: $(target)
	});
}


function _requestShow(tip, text, target, check) {
	_clear(tip);
	if (text) {
		tip.target = target;
		target.syraTipTitle = target.title;
		target.title = "";
		if (!check || (target.syraShortCutTip || target.syraTip == 2 || ((target.scrollWidth - target.clientWidth) > 1))) {
			tip.showTimeOut = setTimeout(function() {
				_show(tip, text, target);
			}, 800);
		}
	}
}

function _clear(tip) {
	if (tip.target) {
		if (tip.target.syraTipTitle) {
			tip.target.title = tip.target.syraTipTitle;
		}
		delete tip.target;
	}
	_hide(tip);
}

function _hide(tip) {
	tip.showTimeOut && clearTimeout(tip.showTimeOut);
	syra_dom.removeChild(tip._item);
	tip.showTimeOut = null;
}

exports.titleTip = {
	$css: "s-title-tip",
	show: function(text, target, check) {
		_requestShow(this, text, target, check);
	},
	hide: function() {
		_hide(this);
	},
	clear: function() {
		_clear(this);
	}
};


exports.fieldTip = {
	$css: "s-field-tip",
	show: function(text, target, check) {
		_requestShow(this, text, target, check);
	},
	onKeyUp: function(field, event) {
		if (!field.isFieldTipDisabled) {
			var target = event.target;
			var value = target && target.value;
			if (this.hasMaxLengthWarn && field.input && value !== undefined) {
				if (field.$field.$maxLength && !(value && value.length >= field.$field.$maxLength)) {
					_clear(this);
					delete this.hasMaxLengthWarn;
				}
			}
		}
	},
	onKeyPress: function(field, event, charCode) {
		if (!field.isFieldTipDisabled) {
			var target = event.target;
			var value = target && target.value;
			if (field.input && target && value !== undefined) {
				if (field.$field.$maxLength) {
					if (charCode && value && value.length >= field.$field.$maxLength) {
						this.hasMaxLengthWarn = true;
						_show(this, syra_local.field_warn_limit_input, this.target = target);
					} else {
						if (this.hasMaxLengthWarn) {
							_clear(this);
							delete this.hasMaxLengthWarn;
						}
					}
				}
			}
		}
	},
	onItemInOut: function(field, on) {
		if (!field.isFieldTipDisabled) {
			if (on) {
				if (field.input) {
					this.show(field.getInputValue(), field.input, true);
				} else {
					this.show(field.displayText || field._dataValue.textContent, field._dataValue, true);
				}
			} else {
				_clear(this);
			}
		}
	}
};
