// syracuse-ui/lib/authoring/storage/pageStorage.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var RawPage = require("syracuse-ui/lib/page/rawPage").RawPage;

function PageStorage() {}

exports.PageStorage = helpers.defineClass(PageStorage, RawPage, {
	load: function(designer) {
		this.isSiteRegisterDisabled = true;
		this.designer = designer;
		this.crudSlot = document.createElement("div");
		this.crudSlot.className = "s-aw-page-wc";
		designer.layoutSlot.appendChild(this.crudSlot);

		this.diagnosePage = designer.designedArticle.page;
		syra_site.pageLoader.initialize(this);
		if (this.designer.designedArticle == this.designer.designedArticle.page && this.designer.designedArticle.page.$authorUrl) {
			this.hasWorkingCopy = true;
			syra_controller.loadWorkingCopy({
				menu: {
					$url: this.designer.designedArticle.page.$authorUrl
				},
				article: this,
				callback: function() {
					syra_site.resize();
				}
			});
		} else {
			this.crudSlot.syraItem = this.crudSlot.syraarticle = this.designer.id;
			this.crudSlot.appendChild(this.designer.addCloseButton());
		}
	},
	onBeforeMainPageChange: function(continueChanging) {
		var self = this;
		if (self.designer.historyTool.isUpdated) {
			syra_site.showMessage({
				$title: syra_local.aw_updateMessageTitle,
				$message: syra_local.aw_cancelMessageText,
				$type: "question",
				$buttons: "yesnocancel",
				callback: function(response) {
					if (response.$clientId == "yes") {
						self.saveUpdate();
						setTimeout(function() {
							self.designer.closeDesigner();
							continueChanging();
						}, 300);
					} else {
						if (response.$clientId == "no") {
							self.designer.closeDesigner();
							continueChanging();
						}
					}
				}
			});
			return false;
		}
		return true;
	},
	onMenuClick: function(menuItem) {
		var self = this;
		switch (menuItem.$item.$bind) {
			case "$delete":
				if (menuItem.page != self.designer.designedArticle.page) {
					if (self.designer.designedArticle.page.pageViewSelector) {
						var selectedView = self.designer.designedArticle.page.pageViewSelector.getSelected();
						syra_site.showMessage({
							$title: syra_local.aw_deleteMessageTitle,
							$message: syra_local.aw_deleteMessageText.replace("{0}", "'" + selectedView.$title + "'"),
							$type: "question",
							$buttons: "yesno",
							callback: function(response) {
								if (response.$clientId == "yes") {
									syra_controller.executeMenu(menuItem);
								}
							}
						});
					}
				}
				return false;
			case "$close":
				if (self.designer.historyTool.isUpdated) {
					syra_site.showMessage({
						$title: syra_local.aw_updateMessageTitle,
						$message: syra_local.aw_cancelMessageText,
						$type: "question",
						$buttons: "yesnocancel",
						callback: function(response) {
							if (response.$clientId == "yes") {
								self.saveUpdate();
							} else {
								if (response.$clientId == "no") {
									self.cancelUpdate();
								}
							}
						}
					});
				} else {
					self.designer.closeDesigner();
				}
				return false;
			default:
				if (menuItem.$sourceBind == "$aw_quit") {
					self.designer.showTargetPageDiagnoses(null);
					self.designer.closeDesigner();
					return false;
				}
				if (menuItem.page != self && !menuItem.page.isMessageBox && !menuItem.boxParent.isDiagnoseMenus && menuItem.page.isDesigned) {
					return false;
				}
		}
		return true;
	},
	saveUpdate: function() {
		syra_menus.clickMenu(this, (this.$menus.$save && !this.$menus.$save.$isDisabled) ? "$save" : "saveAs");
	},
	cancelUpdate: function() {
		var $item = helpers.object.clone(this.designer.historyTool.$sourceItem, true);
		this.designer.designedArticle.page.reloadLayout($item);
		this.designer.applyChangeToMobile($item);
		this.designer.closeDesigner();
	},
	drawBox: function() {
		this.$prototype.$properties = this.$prototype.$properties || {};
		var row = document.createElement("div");
		row.className = "s-aw-menus-row";
		this.crudSlot.appendChild(row);
		this.loadNewItem(row, {
			$bind: "$save",
			$category: "link",
			$skin: "s-aw-menus-save",
			$icon: {
				$mode: "icon",
				$path: "authoring/s-aw-"
			}
		});
		if (!syra_site.mobileGateway) {
			this.loadNewItem(row, {
				$bind: "saveAs",
				$category: "link",
				$skin: "s-aw-menus-saveas",
				$icon: {
					$mode: "icon",
					$path: "authoring/s-aw-"
				}
			});
			this.loadNewItem(row, {
				$bind: "$delete",
				$category: "link",
				$skin: "s-aw-menus-delete",
				$icon: {
					$mode: "icon",
					$path: "authoring/s-aw-"
				}
			});
		}
		this.loadNewItem(row, {
			$bind: "$close",
			$category: "link",
			$skin: "s-aw-page-close",
			$icon: {
				$mode: "icon",
				$path: "authoring/s-aw-"
			}
		});
		this.menuItems.$close[0].setMenu({
			$title: syra_local.aw_close
		}, null);
	},

	onAfterActionMenuExecute: function(menuItem, $menu) {
		switch (menuItem.$bind) {
			case "$save":
			case "saveAs":
				if (this.isActionSuccess(menuItem)) {
					if (this.dataset.$authorUrl) {
						this.designer.designedArticle.page.$authorUrl = this.dataset.$authorUrl;
					}
					this.designer.historyTool.notifyUpdate(false);
					$menu.$diagnoses = [{
						$message: syra_local.aw_saveMessageText,
						$severity: "success"
					}];
					$menu.$links = {
						$create: {
							$isHidden: true
						},
						$details: {
							$isHidden: true
						},
						$query: {
							$isHidden: true
						}
						/*,
                     $aw_quit: {
                     $title: syra_local.aw_quit
                     }*/
					};
					syra_site.deltaManager.applyObjectDelta(menuItem.page, menuItem.articleParent.$menus[menuItem.$bind], {
						$links: $menu.$links
					});
					menuItem.disable(false);
					if (menuItem.$bind == "saveAs") {
						this.designer.designedArticle.page.loadSelectedDesignView(menuItem.$links && menuItem.$links.$view);
					}
				} else {
					if (menuItem.$bind == "saveAs") {
						menuItem.disable(false);
					}
				}
				break;
			case "$delete":
				if (this.isActionSuccess(menuItem)) {
					var selectedView = this.designer.designedArticle.page.pageViewSelector.getSelected();
					if (this.dataset.$authorUrl) {
						this.designer.designedArticle.page.$authorUrl = this.dataset.$authorUrl;
					}
					var $diagnose = {
						$message: syra_local.aw_isDeletedMessageText.replace("{0}", "'" + selectedView.$title + "'"),
						$severity: "success"
					};
					var page = this.designer.designedArticle.page;
					this.designer.closeDesigner();
					page.loadSelectedDesignView();
					return false;
				}
				break;
		}
	},
	applyChange: function(newData) {
		if (newData) {
			newData.$diagnoses = null;
		}
		RawPage.prototype.applyChange.call(this, newData);
	},
	notifyActionChange: function(menuItem, target, value, notifyServer) {
		if (value && value.$save || value.saveAs) {
			this.designer.designedArticle.page.garbage.garbageFreeItems();
		}
		var $article = helpers.object.clone(this.designer.designedArticle.page.$item, true);
		delete $article.$menus;
		delete $article.$isModel;
		if (this.designer.designedArticle.page.makeArticlePersistent) {
			this.designer.designedArticle.page.makeArticlePersistent($article);
		}
		this.saveDesign($article);
		RawPage.prototype.notifyActionChange.call(this, menuItem, target, value, notifyServer);
	},
	saveDesign: function($article) {
		if (this.hasWorkingCopy) {
			this.ensureSendBag().content = {
				$article: syra_site.layoutUpdater.cleanBeforeSave($article || this.designer.designedArticle.page.$item)
			};
		}
	},
	dispose: function() {
		if (this.crudSlot) {
			syra_site.dom.removeChild(this.crudSlot);
		}
		this.crudSlot = this.designer = this.diagnosePage = null;
		RawPage.prototype.dispose.call(this);
	}
});
