// syracuse-ui/lib/tests/pages/field/scenarios.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var _rtf = require("syracuse-rtf");

function _applyMetaToFields(page, key, value) {
	var $properties = {};
	Object.keys(page.boundFields).forEach(function(bind) {
		($properties[bind] = {})[key] = value;
	});
	page.applyChange({
		$properties: $properties
	});
}

function _applyDesignMetaToFields(page, key, value) {
	Object.keys(page.boundFields).forEach(function(bind) {
		page.boundFields[bind].forEach(function(field) {
			var metaData = {};
			metaData[key] = value;
			field.applyDesignMetaData(metaData, true);
		});
	});
}

function _notifyActionLinkChange(page, metaData) {
	var $properties = {};
	Object.keys(page.boundFields).forEach(function(bind) {
		$properties[bind] = {
			$links: metaData.$links,
			$actions: metaData.$actions
		};
	});
	page.applyChange({
		$properties: $properties
	});
}

function _notifyDiagnoses(page, metaData) {
	var $properties = {};
	Object.keys(page.boundFields).forEach(function($bind) {
		var metaData = {};
		var $fieldProperty = page.$prototype.$properties[$bind];
		$properties[$bind] = helpers.object.clone(metaData, true);
	});
	page.applyChange({
		$properties: $properties
	});
}


exports.scenarios = {
	$isHidden: function() {
		return [{
			$title: "$isHidden = true",
			run: function(page) {
				_applyMetaToFields(page, "$isHidden", true);
				//this.assert(first.domItem.querySelector("[data-s-q-bind=xstring]").className.indexOf("s-focus") >= 0);
			}
		}, {
			$title: "$isHidden = false",
			run: function(page) {
				//page.refreshData();
				_applyMetaToFields(page, "$isHidden", false);
				//this.assert(first.domItem.querySelector("[data-s-q-bind=xstring]").className.indexOf("s-focus") >= 0);
			}
		}];
	},
	$isDisabled: function() {
		return [{
			$title: "$isDisabled = true",
			run: function(page) {
				//page.refreshData();
				_applyMetaToFields(page, "$isDisabled", true);
				//this.assert(first.domItem.querySelector("[data-s-q-bind=xstring]").className.indexOf("s-focus") >= 0);
			}
		}, {
			$title: "$isDisabled = false",
			run: function(page) {
				//page.refreshData();
				_applyMetaToFields(page, "$isDisabled", false);
				//this.assert(first.domItem.querySelector("[data-s-q-bind=xstring]").className.indexOf("s-focus") >= 0);
			}
		}];
	},
	$isMandatory: function() {
		return [{
			$title: "$isMandatory = true",
			run: function(page) {
				//page.refreshData();
				_applyMetaToFields(page, "$isMandatory", true);
				//this.assert(first.domItem.querySelector("[data-s-q-bind=xstring]").className.indexOf("s-focus") >= 0);
			}
		}, {
			$title: "$isMandatory = false",
			run: function(page) {
				//page.refreshData();
				_applyMetaToFields(page, "$isMandatory", false);
				//this.assert(first.domItem.querySelector("[data-s-q-bind=xstring]").className.indexOf("s-focus") >= 0);
			}
		}];
	},
	$isEditMode: function() {
		return [{
			$title: "$isEditMode = true",
			run: function(page) {
				//page.refreshData();
				_applyMetaToFields(page, "$isEditMode", true);
				//this.assert(first.domItem.querySelector("[data-s-q-bind=xstring]").className.indexOf("s-focus") >= 0);
			}
		}, {
			$title: "$isEditMode = false",
			run: function(page) {
				//page.refreshData();
				_applyMetaToFields(page, "$isEditMode", false);
				//this.assert(first.domItem.querySelector("[data-s-q-bind=xstring]").className.indexOf("s-focus") >= 0);
			}
		}];
	},
	$isReadOnly: function() {
		return [{
			$title: "$isReadOnly = true",
			run: function(page) {
				//page.refreshData();
				_applyMetaToFields(page, "$isReadOnly", true);
				//this.assert(first.domItem.querySelector("[data-s-q-bind=xstring]").className.indexOf("s-focus") >= 0);
			}
		}, {
			$title: "$isReadOnly = false",
			run: function(page) {
				//page.refreshData();
				_applyMetaToFields(page, "$isReadOnly", false);
				//this.assert(first.domItem.querySelector("[data-s-q-bind=xstring]").className.indexOf("s-focus") >= 0);
			}
		}];
	},
	$description: function() {
		return [{
			$title: "$description = 'this is a description'",
			run: function(page) {
				//page.refreshData();
				_applyMetaToFields(page, "$description", "this is a description");
				//this.assert(first.domItem.querySelector("[data-s-q-bind=xstring]").className.indexOf("s-focus") >= 0);
			}
		}, {
			$title: "$description = null",
			run: function(page) {
				//page.refreshData();
				_applyMetaToFields(page, "$description", null);
				//this.assert(first.domItem.querySelector("[data-s-q-bind=xstring]").className.indexOf("s-focus") >= 0);
			}
		}];
	},
	$isTopLabelAlignment: function() {
		return [{
			$title: "$isTopLabelAlignment = true",
			run: function(page) {
				//page.refreshData();
				_applyDesignMetaToFields(page, "$isTopLabelAlignment", true);
				//this.assert(first.domItem.querySelector("[data-s-q-bind=xstring]").className.indexOf("s-focus") >= 0);
			}
		}, {
			$title: "$isTopLabelAlignment = false",
			run: function(page) {
				//page.refreshData();
				_applyDesignMetaToFields(page, "$isTopLabelAlignment", false);
				//this.assert(first.domItem.querySelector("[data-s-q-bind=xstring]").className.indexOf("s-focus") >= 0);
			}
		}];
	},
	$isTitleHidden: function() {
		return [{
			$title: "$isTitleHidden = true",
			run: function(page) {
				//page.refreshData();
				_applyDesignMetaToFields(page, "$isTitleHidden", true);
				//this.assert(first.domItem.querySelector("[data-s-q-bind=xstring]").className.indexOf("s-focus") >= 0);
			}
		}, {
			$title: "$isTitleHidden = false",
			run: function(page) {
				//page.refreshData();
				_applyDesignMetaToFields(page, "$isTitleHidden", false);
				//this.assert(first.domItem.querySelector("[data-s-q-bind=xstring]").className.indexOf("s-focus") >= 0);
			}
		}];
	},
	setNull: function() {
		return [{
			$title: "set null",
			run: function(page) {
				//page.refreshData();
				Object.keys(page.boundFields).forEach(function(bind) {
					page.boundFields[bind].forEach(function(field) {
						field.setDataBind(null);
					});
				});
				//this.assert(first.domItem.querySelector("[data-s-q-bind=xstring]").className.indexOf("s-focus") >= 0);
			}
		}];
	},
	actionLinks: function() {
		return [{
			$title: "$links = null",
			run: function(page) {
				//page.refreshData();
				_notifyActionLinkChange(page, {
					$links: null,
					$actions: null
				});
				//this.assert(first.domItem.querySelector("[data-s-q-bind=xstring]").className.indexOf("s-focus") >= 0);
			}
		}, {
			$title: "$lazyload (transition)",
			run: function(page) {
				//page.refreshData();
				_notifyActionLinkChange(page, {
					$links: {
						$lazyload: {}
					}
				});
				//this.assert(first.domItem.querySelector("[data-s-q-bind=xstring]").className.indexOf("s-focus") >= 0);
			}
		}, {
			$title: "links = {}",
			run: function(page) {
				//page.refreshData();
				_notifyActionLinkChange(page, {
					$links: {}
				});
				//this.assert(first.domItem.querySelector("[data-s-q-bind=xstring]").className.indexOf("s-focus") >= 0);
			}
		}, {
			$title: "$links and $actions",
			run: function(page) {
				//page.refreshData();
				_notifyActionLinkChange(page, {
					$links: {
						"$tunnel": {
							"$title": "$tunnel"
						},
						"$lookup": {
							"$title": "$lookup"
						},
						"google": {
							"$title": "Google",
							"$url": "http://www.google.fr/"
						},
						"uioverview": {
							"$title": "UI Overview",
							"$url": "?representation=s-uitest-fields.$test"
						},
						"$create": {
							"$type": "application/json;vnd.sage=syracuse",
							"$title": "Créer",
							"$url": "http://localhost:8126/sdata/syracuse/collaboration/syracuse/users/$template/$workingCopies?representation=country.$edit&role={$role}",
							"$method": "POST"
						},
						"$details": {
							"$type": "application/json;vnd.sage=syracuse",
							"$title": "Detail",
							"$url": "http://localhost:8126/sdata/syracuse/collaboration/syracuse/users('{$uuid}')?representation=country.$details&role={$role}"
						},
						"$edit": {
							"$type": "application/json;vnd.sage=syracuse",
							"$title": "Edit",
							"$url": "http://localhost:8126/sdata/syracuse/collaboration/syracuse/users('{$uuid}')/$workingCopies?representation=country.$edit&role={$role}",
							"$method": "POST"
						},
						"$delete": {
							"$title": "Delete",
							"$confirm": "The record  will be deleted. Confirm ?",
							"$url": "http://localhost:8126/sdata/syracuse/collaboration/syracuse/users('{$uuid}')?representation=country.$query&role={$role}",
							"$type": "application/json;vnd.sage=syracuse",
							"$method": "DELETE"
						}
					},
					$actions: {
						actionEmployeesProperties: {
							"$title": "Action Employees Properties",
							"$url": "/sdata/syracuse/collaboration/syracuse/users?representation=user.$query",
							"$parameters": {
								"param1": "{lastName}",
								"param2": "value"
							}
						}
					}
				});
				//this.assert(first.domItem.querySelector("[data-s-q-bind=xstring]").className.indexOf("s-focus") >= 0);
			}
		}];
	},
	$diagnoses: function() {
		return [{
			$title: "$diagnoses show error, warning, info",
			run: function(page) {
				//page.refreshData();
				_notifyDiagnoses(page, {
					$diagnoses: [{
						severity: "error",
						message: "Error test  blalfoz"
					}]
				});
				//this.assert(first.domItem.querySelector("[data-s-q-bind=xstring]").className.indexOf("s-focus") >= 0);
			}
		}, {
			$title: "$diagnoses show error",
			run: function(page) {
				//page.refreshData();
				_notifyDiagnoses(page, {
					$diagnoses: [{
						severity: "error",
						message: "Error test  blalfoz"
					}, {
						severity: "warning",
						message: "Warning test"
					}, {
						severity: "info",
						message: "Info test"
					}]
				});
				//this.assert(first.domItem.querySelector("[data-s-q-bind=xstring]").className.indexOf("s-focus") >= 0);
			}
		}, {
			$title: "$diagnoses = null",
			run: function(page) {
				//page.refreshData();
				_notifyDiagnoses(page, {
					$diagnoses: null
				});
				//this.assert(first.domItem.querySelector("[data-s-q-bind=xstring]").className.indexOf("s-focus") >= 0);
			}
		}];
	}
};


exports.editor = function() {
	return [{
		$title: "convert HTML to RTF",
		run: function(page) {
			//page.refreshData();
			page.boundFields["textplainRTF"][0].setDataValue(rtf.fromHtml($(page.boundFields["textrtf"][0]._editor.doc.body).html()));
			//this.assert(first.domItem.querySelector("[data-s-q-bind=xstring]").className.indexOf("s-focus") >= 0);
		}
	}, {
		$title: "convert RTF to HTML",
		run: function(page) {
			//page.refreshData();
			$(page.boundFields["textrtf"][0]._editor.doc.body).html(rtf.toHtml(page.boundFields["textplainRTF"][0].getDataValue()));
			//this.assert(first.domItem.querySelector("[data-s-q-bind=xstring]").className.indexOf("s-focus") >= 0);
		}
	}];
};



exports.notifyDataChange = function(field, value) {
	var rtfField = field.page.boundFields["textrtf"][0],
		msg = "textrtf: caret=" + rtfField.getCaretPosition() + ", data=" + rtfField.getInputValue();
	field.page.boundFields["$field$output"][0].setDataValue(msg);
};
