// syracuse-ui/lib/tests/pages/field/scenarios.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var _rtf = require("syracuse-rtf");

function _applyMeta(unit, key, value, strictEqual) {
	var $properties = {};
	Object.keys(unit.page.boundFields).forEach(function(bind) {
		($properties[bind] = {})[key] = value;
	});
	unit.applyChange({
		$properties: $properties
	});
	var fields = syra_quality.domQuery.getFields([syra_quality.domQuery.page.getBody(unit.page.domItem)]);
	var ok;
	var successCount = 0,
		errorCount = 0;
	for (var mm = 0; mm < fields.length; mm++) {
		var dom = fields[mm];
		var ok = strictEqual(dom, syra_quality.domQuery.field.getBind(dom), syra_quality.domQuery.field.getType(dom));
		if (ok !== undefined) {
			var title = syra_quality.domQuery.field.getTitle(dom);
			var editmode = syra_quality.domQuery.field.isEditMode(dom) ? " editmode " : " ";
			ok ? successCount++ : errorCount++;
			if (!unit.page.isAssertSummaryMode) {
				unit.assert(ok, "fied " + title + editmode + unit.$title);
			}
		}
	}
	if (unit.page.isAssertSummaryMode) {
		successCount && unit.assert(true, successCount + " fied " + title + editmode + unit.$title);
		errorCount && unit.assert(false, errorCount + " fied " + title + editmode + unit.$title);
	}
}

function _applyDesignMetaToFields(unit, key, value, strictEqual) {
	Object.keys(unit.page.boundFields).forEach(function(bind) {
		unit.page.boundFields[bind].forEach(function(field) {
			var metaData = {};
			metaData[key] = value;
			field.applyDesignMetaData(metaData, true);
			unit.assert(strictEqual(field), "fied " + field.getTitle() + (field.isEditMode ? " editmode " : " ") + unit.$title);
		});
	});

}

function _notifyActionLinkChange(page, metaData) {
	var $properties = {};
	Object.keys(page.boundFields).forEach(function(bind) {
		$properties[bind] = {
			$links: metaData.$links,
			$actions: metaData.$actions
		};
	});
	page.applyChange({
		$properties: $properties
	});
}

function _addStatusUnits(status) {
	return [true, false].map(function(isEnabled) {
		return {
			$title: status + " = " + isEnabled,
			run: function() {
				_applyMeta(this, status, isEnabled, function(dom, bind, type) {
					var ok;
					if (type.indexOf("x-unknow") < 0) {
						var isEditMode = syra_quality.domQuery.field.isEditMode(dom);
						switch (status) {
							case "$isHidden":
								ok = (syra_quality.domQuery.field.isHidden(dom) == isEnabled);
								break;
							case "$isDisabled":
								if (isEditMode) {
									ok = (syra_quality.domQuery.field.isInputDisabled(dom, true) == isEnabled);
								} else {
									ok = true;
								}
								break;
							case "$isReadOnly":
								if (isEditMode) {
									ok = (syra_quality.domQuery.field.isInputReadOnly(dom, true) == isEnabled);
								} else {
									ok = true;
								}
								break;
							case "$isMandatory":
								if (isEditMode && type.indexOf("x-boolean") < 0) {
									ok = (syra_quality.domQuery.field.isMandatory(dom, true) == isEnabled);
								} else {
									ok = true;
								}
								break;
						}
					}
					return ok;
				});
			}
		};
	});
}

exports.scenarios = {
	$isHidden: function() {
		return _addStatusUnits("$isHidden");
	},
	$isDisabled: function() {
		return _addStatusUnits("$isDisabled");
	},
	$isReadOnly: function() {
		return _addStatusUnits("$isReadOnly");
	},
	$isMandatory: function() {
		return _addStatusUnits("$isMandatory");
	},
	/*$isEditMode: function(){
     return _addStatusUnits("$isEditMode");
     },*/
	$description: function() {
		return ["this is a description", null].map(function(value) {
			return {
				$title: "$description = " + (value ? ("'" + value + "'") : "null"),
				run: function() {
					_applyMeta(this, "$description", value, function(dom, bind, type) {
						if (type.indexOf("x-unknow") < 0) {
							var desc = syra_quality.domQuery.field.findDescription(dom);
							if (desc) {
								var ok = (desc.innerText == value) &&
									(desc.getAttribute("data-s-description") == value);
								return ok;
							} else {
								if (value) {
									var field = syra_store.findField(dom);
									if (field.$item.$inplace) {
										return true; //no description for inplace 	
									}
								}
							}
							return !value;
						}
					});
				}
			};
		});
	},
	$isTopLabelAlignment: function() {
		return [true, false].map(function(value) {
			return {
				$title: "$isTopLabelAlignment = " + value,
				run: function(page) {
					_applyDesignMetaToFields(this, "$isTopLabelAlignment", value, function(field) {
						var ok = field.$item.$isTopLabelAlignment == value;
						if (ok && field.domTitle) {
							var ii = field.domTitle.className.indexOf("s-beside");
							ok = ((ii == -1 && value) || (!value && ii >= 0));
						}
						return ok;
					});
				}
			};
		});
	},
	$isTitleHidden: function() {
		return [true, false].map(function(value) {
			return {
				$title: "$isTitleHidden = " + value,
				run: function(page) {
					_applyDesignMetaToFields(this, "$isTitleHidden", value, function(field) {
						var ok = field.$item.$isTitleHidden == value;
						if (ok && field.domTitle) {
							ok = field.domTitle.style.display == (value ? "none" : "");
						}
						return ok;
					});
				}
			};
		});
	},
	/*setNull: function() {
		return [{
			$title: "set null",
			run: function(page) {
				var self = this;
				var data = {};
				Object.keys(page.boundFields).forEach(function(bind) {
					data[bind] = null;
				});
				self.applyChange(data);
				var fields = syra_quality.domQuery.getFields([syra_quality.domQuery.page.getBody(self.page.domItem)]);
				var ok;
				for (var mm = 0; mm < fields.length; mm++) {
					var dom = fields[mm];
					var field = syra_store.findField(dom);
					var type = syra_quality.domQuery.field.getType(dom);
					if (type.indexOf("x-unknow") < 0 && type.indexOf("quantity") < 0) {
						var title = syra_quality.domQuery.field.getTitle(dom);
						var editmode = syra_quality.domQuery.field.isEditMode(dom) ? " editmode " : " ";
						var value = syra_quality.domQuery.field.getDataValue(dom);
						var ok = !value;
						if (!ok) {
							if ((field.$numeric && value == "0") ||
								(field.isBooleanField && !field.currentValue)) {
								ok = true;
							}
						}
						if (ok && field.currentValue) {
							ok = false;
							if (field.isReferenceField) {
								ok = Object.keys(field.currentValue).length == 0;
							}
						}
						self.assert(ok, "fied " + title + editmode + self.$title);
					}
				}
			}
		}];
	},*/
	actionLinks: function() {
		return [{
			$title: "$links = null",
			run: function(page) {
				//page.refreshData();
				_notifyActionLinkChange(page, {
					$links: null,
					$actions: null
				});
				//this.assert(first.domItem.querySelector("[data-s-bind=xstring]").className.indexOf("s-focus") >= 0);
			}
		}, {
			$title: "$lazyload (transition)",
			run: function(page) {
				//page.refreshData();
				_notifyActionLinkChange(page, {
					$links: {
						$lazyload: {}
					}
				});
				//this.assert(first.domItem.querySelector("[data-s-bind=xstring]").className.indexOf("s-focus") >= 0);
			}
		}, {
			$title: "links = {}",
			run: function(page) {
				//page.refreshData();
				_notifyActionLinkChange(page, {
					$links: {}
				});
				//this.assert(first.domItem.querySelector("[data-s-bind=xstring]").className.indexOf("s-focus") >= 0);
			}
		}, {
			$title: "$links and $actions",
			run: function(page) {
				//page.refreshData();
				_notifyActionLinkChange(page, {
					$links: {
						"$tunnel": {
							"$title": "$tunnel"
						},
						"$lookup": {
							"$title": "$lookup"
						},
						"google": {
							"$title": "Google",
							"$url": "http://www.google.fr/"
						},
						"uioverview": {
							"$title": "UI Overview",
							"$url": "?representation=s-uitest-fields.$test"
						},
						"$create": {
							"$type": "application/json;vnd.sage=syracuse",
							"$title": "Créer",
							"$url": "http://localhost:8126/sdata/syracuse/collaboration/syracuse/users/$template/$workingCopies?representation=country.$edit&role={$role}",
							"$method": "POST"
						},
						"$details": {
							"$type": "application/json;vnd.sage=syracuse",
							"$title": "Detail",
							"$url": "http://localhost:8126/sdata/syracuse/collaboration/syracuse/users('{$uuid}')?representation=country.$details&role={$role}"
						},
						"$edit": {
							"$type": "application/json;vnd.sage=syracuse",
							"$title": "Edit",
							"$url": "http://localhost:8126/sdata/syracuse/collaboration/syracuse/users('{$uuid}')/$workingCopies?representation=country.$edit&role={$role}",
							"$method": "POST"
						},
						"$delete": {
							"$title": "Delete",
							"$confirm": "The record  will be deleted. Confirm ?",
							"$url": "http://localhost:8126/sdata/syracuse/collaboration/syracuse/users('{$uuid}')?representation=country.$query&role={$role}",
							"$type": "application/json;vnd.sage=syracuse",
							"$method": "DELETE"
						}
					},
					$actions: {
						actionEmployeesProperties: {
							"$title": "Action Employees Properties",
							"$url": "/sdata/syracuse/collaboration/syracuse/users?representation=user.$query",
							"$parameters": {
								"param1": "{lastName}",
								"param2": "value"
							}
						}
					}
				});
				//this.assert(first.domItem.querySelector("[data-s-bind=xstring]").className.indexOf("s-focus") >= 0);
			}
		}];
	},
	$diagnoses: function() {
		function _notifyDiagnoses(page, metaData) {
			var $properties = {};
			Object.keys(page.boundFields).forEach(function($bind) {
				$properties[$bind] = helpers.object.clone(metaData, true);
			});
			page.applyChange({
				$properties: $properties
			});
		}
		return [{
			$title: "$diagnoses show error, warning, info",
			run: function(page) {
				//page.refreshData();
				_notifyDiagnoses(page, {
					$diagnoses: [{
						severity: "error",
						message: "Error test  blalfoz"
					}]
				});
				//this.assert(first.domItem.querySelector("[data-s-bind=xstring]").className.indexOf("s-focus") >= 0);
			}
		}, {
			$title: "$diagnoses show error",
			run: function(page) {
				//page.refreshData();
				_notifyDiagnoses(page, {
					$diagnoses: [{
						severity: "error",
						message: "Error test  blalfoz"
					}, {
						severity: "warning",
						message: "Warning test"
					}, {
						severity: "info",
						message: "Info test"
					}]
				});
				//this.assert(first.domItem.querySelector("[data-s-bind=xstring]").className.indexOf("s-focus") >= 0);
			}
		}, {
			$title: "$diagnoses = null",
			run: function(page) {
				//page.refreshData();
				_notifyDiagnoses(page, {
					$diagnoses: null
				});
				//this.assert(first.domItem.querySelector("[data-s-bind=xstring]").className.indexOf("s-focus") >= 0);
			}
		}];
	}
};


exports.editor = function() {
	return [{
		$title: "convert HTML to RTF",
		run: function(page) {
			//page.refreshData();
			page.boundFields["textplainRTF"][0].setDataValue(rtf.fromHtml($(page.boundFields["textrtf"][0]._editor.doc.body).html()));
			//this.assert(first.domItem.querySelector("[data-s-bind=xstring]").className.indexOf("s-focus") >= 0);
		}
	}, {
		$title: "convert RTF to HTML",
		run: function(page) {
			//page.refreshData();
			$(page.boundFields["textrtf"][0]._editor.doc.body).html(rtf.toHtml(page.boundFields["textplainRTF"][0].getDataValue()));
			//this.assert(first.domItem.querySelector("[data-s-bind=xstring]").className.indexOf("s-focus") >= 0);
		}
	}];
};



exports.notifyDataChange = function(field, value) {
	var rtfField = field.page.boundFields["textrtf"][0],
		msg = "textrtf: caret=" + rtfField.getCaretPosition() + ", data=" + rtfField.getInputValue();
	field.page.boundFields["$field$output"][0].setDataValue(msg);
};
