// syracuse-ui/lib/authoring/aside/rowResizer.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function RowResizer(){
}

exports.RowResizer = helpers.defineClass(RowResizer, null, {
    load: function(){
        var self = this;
        self.percents = [16, 20, 25, 33, 40, 50, 60, 66, 75, 80, 84];
        document.site.$$layoutSlot.delegate(".s-aw-slot-row-sep", "mousedown.syraauthorresizer", function(event){
            self.start(this, event);
            return false;
        });
    },
    start: function(domSeparator, event){
        var self = this;
        self.onResizing = true;
        var authorPage = document.site.authorPage;
        self.startX = event.pageX;
        self.index = parseInt(domSeparator.getAttribute("data-s-index"), 10);
        self.layout = authorPage.findTargetItemEvent(domSeparator, true);
        self.widths = authorPage.layoutValidator.getWidthValues(self.layout.$layout.$widths)
        
        debugger;
        self.leftItem = self.layout.items[self.index - 1];
        self.leftWidth = $(self.leftItem.layoutSlot).width();
        self.rightItem = self.layout.items[self.index];
        self.rightWidth = $(self.rightItem.layoutSlot).width();
        self.offset = $(self.layout.$$item).offset();
        
        self.steps = [];
        var max = self.maxWidth = $(self.layout.$$item).width();
        var steps = [max / 6, max / 5, max / 4, max / 3, 2 * (max / 5), max / 2, 3 * (max / 5), 2 * (max / 3), 3 * (max / 4), 4 * (max / 5), 5 * (max / 6)];
        self.margin = document.site.$$body.offset();
        self.$$rowSeparator = $(domSeparator);
        
        var minWidth = 0, maxWidth = 0;
        for (var ii = 0, jj = self.widths.length; ii < jj; ii++) {
            if (ii < (self.index - 1)) {
                minWidth += self.widths[ii];
            }
            if (ii <= (self.index)) {
                maxWidth += self.widths[ii];
            }
        }
        // var minWidth = Math.max(self.layout.items[Math.max(self.index - 2, 0)], 16);
        //var maxWidth = Math.max(self.layout.items[Math.min(self.index + 1, self.widths.length - 1)], 16);
        
        minWidth = Math.max(self.percents[0], minWidth);
        maxWidth = Math.min(self.percents[self.percents.length - 1], maxWidth);
        for (var ii = 0, jj = steps.length; ii < jj; ii++) {
            var width = steps[ii];
            var marker = document.createElement("div");
            marker.className = "s-aw-resizer-marker";
            var pageX = self.offset.left + width;
            marker.style.left = (pageX - self.margin.left) + "px";
            marker.style.top = (self.offset.top - self.margin.top - 10) + "px";
            self.steps.push({
                pageX: pageX,
                offsetX: width,
                width: width,
                marker: document.site.body.appendChild(marker)
            });
            var percent = self.percents[ii];
            if (((percent - 2) <= minWidth) && (minWidth <= (percent + 2))) {
                self.minStep = ii;
            }
            if (((percent - 2) <= maxWidth) && (maxWidth <= (percent + 2))) {
                self.maxStep = ii;
            }
        }
        if (self.index > 1) {
            self.minStep++;
        }
        if (self.index < (self.widths.length - 1)) {
            self.maxStep--;
        }
        document.site.body.style.cursor = "w-resize";
        
        $(document).bind("mouseup.syraauthorresizing", function(event){
            var widths = self.layout.$layout.$widths.split(",");
            for (var ii = 0, jj = self.layout.rowSeparators.length; ii < jj; ii++) {
                var $$rowSeparator = $(self.layout.rowSeparators[ii]);
                /*var separatorOffset = $$rowSeparator.offset();
                 for (var mm = 0, kk = self.steps.length; mm < kk; mm++) {
                 var step = self.steps[mm];
                 if ((step.pageX + 5) > separatorOffset.left) {
                 debugger;
                 widths[ii] = "";
                 }
                 }*/
            }
            self.layout.$layout.$widths = widths.join(",");
            self.stop();
        }).bind("mousemove.syraauthorresizing", function(event){
            self.onSeparatorMove(event)
            event.preventDefault();
        });
    },
    onSeparatorMove: function(event){
        if (event.pageX >= this.steps[this.minStep].pageX && event.pageX <= this.steps[this.maxStep].pageX) {
            var diff = event.pageX - this.startX;
            this.leftItem.layoutSlot.style.width = Math.max(this.leftWidth + diff) + "px";
            this.rightItem.layoutSlot.style.width = this.rightWidth + (diff * (-1)) + "px";
        }
    },
    stop: function(){
        document.site.body.style.cursor = "default";
        $(document).unbind(".syraauthorresizing");
        if (this.steps) {
            for (var ii = 0, jj = this.steps.length; ii < jj; ii++) {
                document.site.removeDomChild(this.steps[ii].marker);
                this.steps[ii].marker = null;
            }
        }
        this.index = this.leftWidth = this.rightWidth = this.offset = null;
        this.steps = this.widths = this.layout = this.leftItem = this.rightItem = null;
        this.onResizing = false;
    },
    dispose: function(){
        this.stop();
        if (document.site.$$layoutSlot) {
            document.site.$$layoutSlot.undelegate(".syraauthorresizer");
        }
    }
});
