// syracuse-ui/lib/authoring/aside/toolsBar.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var RawPage = require("syracuse-ui/lib/page/rawPage").RawPage;
var TreeContent = require("./treeContent").TreeContent;
function ToolsBar(){
}

exports.ToolsBar = helpers.defineClass(ToolsBar, RawPage, {
    loadToolbar: function(awContext){
        this.awContext = awContext;
        this._initializePage();
        this._unregisterResizer = true;
        this.$autoFetch = false;
        this.$item = {};
        this.$prototype = {
            $properties: {},
            $links: {
                $addSection: {
                    $title: "{@addSection}"
                },
                $addBlock: {
                    $title: "{@addBlock}"
                },
                $addMenus: {
                    $title: "{@addMenus}"
                },
                $defaultLayout: {
                    $title: "{@defaultLayout}"
                },
                "pattern-headerTabs": {
                    $title: "{@autoLayout}"
                },
                "pattern-1": {
                    $title: "{@layoutComposite}"
                },
                "pattern-2": {
                    $title: "{@layoutComposite}"
                },
                "pattern-3": {
                    $title: "{@layoutComposite}"
                },
                "pattern-4": {
                    $title: "{@layoutComposite}"
                }
            }
        };
        this.$prototype.$localization = this.awContext.palette.$prototype.$localization;
        
        this.preferences = {
            isDocked: true,
            isCollapsed: this.awContext.targetPage.dialogWrapper
        };
        
        this.$$slot = $(this.layoutSlot = document.createElement("div"));
        this.layoutSlot.className = "s-aw-toolbar-slot";
        this.layoutSlot.setAttribute("data-s-article", this.id);
        this.awContext.targetPage._item.insertBefore(this.layoutSlot, this.awContext.targetPage._item.firstChild);
        
        this.awContext.treeContent = new TreeContent();
        this.awContext.treeContent.load(this.awContext);
        
        this.loadBox();
        
    },
    onEndChangeStep: function(){
        this.reloadFields();
        this.awContext.treeContent.toggle(true);
    },
    dispose: function(){
        if (this._resizer) {
            this._resizer.dispose();
            this._resizer = null;
        }
        this.toggleItemAuthoring(false);
        if (this.blocks) {
            var ids = Object.keys(this.blocks);
            for (var ii = 0, jj = ids.length; ii < jj; ii++) {
                delete this.blocks[ids[ii]].content;
            }
        }
        if (this.layoutSlot) {
            document.site.removeDomChild(this.layoutSlot);
        }
        if (this.$prototype) {
            this.$prototype = this.$prototype.$localization = null;
        }
        if (this.awContext) {
            if (this.awContext.treeContent) {
                this.awContext.treeContent.dispose();
                this.awContext.treeContent = null;
            }
        }
        this.awContext = this.blocks = this.layoutSlot = null;
    },
    _fillInsertMenusBlock: function($actions, $items, $type){
        var $menus = this.awContext.targetPage.$prototype[$type];
        if ($menus) {
            var binds = Object.keys($menus);
            for (var ii = 0, jj = binds.length; ii < jj; ii++) {
                var bind = binds[ii];
                var $field = $menus[bind];
                var actionId = "$fieldLink" + bind;
                var title = $field.$title ? this.awContext.targetPage.getLocalizeText($field.$title) : "";
                $actions[actionId] = {
                    $title: (title == "") ? bind : title
                };
                $items.push({
                    $bind: actionId,
                    $fieldBind: bind,
                    $fieldCategory: "link",
                    $css: "s-aw-add-item",
                    $value: "menu"
                });
            }
        }
    },
    _fillInsertBlock: function(addLinks){
        var $items = [];
        var $actions = {};
        if (addLinks) {
            this._fillInsertMenusBlock($actions, $items, "$links");
            this._fillInsertMenusBlock($actions, $items, "$actions");
        }
        else {
            var binds = Object.keys(this.awContext.targetPage.$prototype.$properties);
            for (var ii = 0, jj = binds.length; ii < jj; ii++) {
                var bind = binds[ii];
                if (bind.charAt(0) != "$") {
                    var $field = this.awContext.targetPage.$prototype.$properties[bind];
                    if ($field && $field.$type) {
                        var actionId = "$fieldLink" + bind;
                        var title = $field.$title ? this.awContext.targetPage.getLocalizeText($field.$title) : "";
                        $actions[actionId] = {
                            $title: (title == "") ? bind : title
                        };
                        $items.push({
                            $bind: actionId,
                            $fieldCategory: "field",
                            $fieldBind: bind,
                            $css: "s-aw-add-item",
                            $value: $field.$type.replace("application/x-", "").replace("/", "-")
                        });
                    }
                }
            }
        }
        
        var fields = this.awContext.drawHelper.extractFields(this.awContext.targetPage.layoutContent);
        for (var ii = 0, jj = fields.length; ii < jj; ii++) {
            var $action = $actions["$fieldLink" + fields[ii].$item.$bind];
            if ($action) {
                $action.$isHidden = true;
            }
        }
        this.applyChange({
            $actions: $actions
        });
        return $items;
    },
    reloadFields: function(){
        var boxIds = ["$menuInsertLinks", "$menuInsertFields"];
        for (var ii = 0, jj = boxIds.length; ii < jj; ii++) {
            var menuBox = this.idMap[boxIds[ii]];
            if (menuBox && menuBox.layoutContent) {
                var items = menuBox.layoutContent.items;
                while (items.length > 0) {
                    menuBox.layoutContent.removeItem(items[0], true);
                }
                this.removeItem(menuBox.layoutContent, true);
            }
        }
        var ids = Object.keys(this.$menus);
        for (var ii = 0, jj = ids.length; ii < jj; ii++) {
            var id = ids[ii];
            if (id.indexOf("$fieldLink") >= 0) {
                delete this.menuItems[id];
                delete this.$menus[id];
            }
        }
        for (var ii = 0, jj = boxIds.length; ii < jj; ii++) {
            var menuBox = this.idMap[boxIds[ii]];
            if (menuBox) {
                menuBox.$item.$layout.$items = this._fillInsertBlock(boxIds[ii] == "$menuInsertLinks");
                menuBox.ensureLinkSettings(menuBox.$item, menuBox.$item.$isCollapsable);
                menuBox.renderLayoutContent();
            }
        }
    },
    drawBox: function(){
        document.site.emptyDom(this.layoutSlot);
        this._bar = document.createElement("div");
        this._bar.className = "s-aw-toolbar";
        this.$$bar = $(this._bar);
        this.appendPickerBar();
        
        this._body = document.createElement("div");
        this._body.className = "s-aw-toolbar-body";
        this._$$body = $(this._body);
        this._bar.appendChild(this._body);
        this.blocks = {};
        
        this.createBlock({
            $category: "section",
            $id: "$treeContentBlock",
            $title: this.awContext.localize.aw_viewTypeContent,
            $layout: {
                $items: []
            }
        });
        var isDashboard = this.awContext.targetPage.isDashBoard;
        this.createBlock({
            $category: "section",
            $id: "$insertBlock",
            $title: this.awContext.localize.aw_insert,
            $layout: {
                $items: [{
                    $category: "menus",
                    $itemIcon: {
                        "$mode": "iconText",
                        "$path": "authoring/s-aw-"
                    },
                    $skin: "s-aw-menus",
                    $itemSkin: "s-aw-menus-link",
                    $layout: {
                        $items: [{
                            $css: "s-aw-add-section s-aw-add-item",
                            $bind: "$addSection",
                            $value: "add-section"
                        }, {
                            $css: "s-aw-add-block s-aw-add-item",
                            $bind: "$addBlock",
                            $value: "add-block",
                            $isHidden: isDashboard
                        }, {
                            $category: "menus",
                            $id: "$menuInsertFields",
                            $itemSkin: "s-aw-mn-field-link",
                            $titleIcon: {
                                "$value": "template",
                                "$path": "authoring/s-aw-"
                            },
                            $itemIcon: {
                                "$mode": "iconText",
                                "$path": "authoring/s-aw-"
                            },
                            $layout: {
                                $items: this._fillInsertBlock()
                            }
                        }]
                    }
                }]
            }
        });
        if (isDashboard) {
            this.createBlock({
                $category: "section",
                $id: "$insertLinks",
                $title: this.awContext.localize.aw_insert_links,
                $layout: {
                    $items: [{
                        $category: "menus",
                        $itemIcon: {
                            "$mode": "iconText",
                            "$path": "authoring/s-aw-"
                        },
                        $skin: "s-aw-menus",
                        $itemSkin: "s-aw-menus-link",
                        $layout: {
                            $items: [{
                                $css: "s-aw-add-section s-aw-add-item",
                                $bind: "$addSection",
                                $value: "add-section"
                            }, {
                                $css: "s-aw-add-block s-aw-add-item",
                                $bind: "$addMenus",
                                $value: "add-menu-group"
                            }, {
                                $category: "menus",
                                $id: "$menuInsertLinks",
                                $itemSkin: "s-aw-mn-field-link",
                                $titleIcon: {
                                    "$value": "template",
                                    "$path": "authoring/s-aw-"
                                },
                                $itemIcon: {
                                    "$mode": "iconText",
                                    "$path": "authoring/s-aw-"
                                },
                                $layout: {
                                    $items: this._fillInsertBlock(true)
                                }
                            }]
                        }
                    }]
                }
            });
        }
        this.createBlock({
            $category: "section",
            $id: "$modelsBlock",
            $title: this.awContext.localize.aw_models,
            $layout: {
                $items: [{
                    $id: "s-awModels",
                    $category: "menus",
                    $itemIcon: {
                        $mode: "icon",
                        $path: "authoring/s-aw-"
                    },
                    $skin: "s-aw-menus",
                    $itemSkin: "s-aw-menus-link",
                    $layout: {
                        $items: [{
                            $icon: null,
                            $bind: "pattern-headerTabs"
                        }, {
                            $icon: null,
                            $bind: "$defaultLayout"
                        }, {
                            $bind: "pattern-1"
                        }, {
                            $bind: "pattern-2"
                        }, {
                            $bind: "pattern-3"
                        }, {
                            $bind: "pattern-4"
                        }]
                    }
                }]
            }
        });
        this.$$bar.appendTo(this.layoutSlot);
        if (!this.openedBlock) {
            this.collapseBlock(this.blocks.$treeContentBlock, true, null, true);
        }
        this.ensureState();
    },
    
    onClickPicker: function(pickerId, event){
        switch (pickerId) {
            case "s-bar-dockMode":
                this.preferences.isDocked = !this.preferences.isDocked;
                break;
            case "s-bar-collapse":
                this._resizer.isEnabled = !(this.preferences.isCollapsed = !this.preferences.isCollapsed);
                break;
            case "s-bar-title":
                var $id = $(event.target).closest("[data-s-block]").attr("data-s-block");
                this.collapseBlock(this.blocks[$id], true);
                break;
        }
        this.ensureState();
        this.onWindowResize();
        this.awContext.palette.resizeDialogPage();
    },
    appendPickerBar: function(){
        var self = this;
        self.resizeBar = document.createElement("a");
        self.resizeBar.className = "s-aw-toolbar-resizer";
        self.openerPicker = document.createElement("a");
        self.openerPicker.className = "s-aw-toolbar-opener";
        self.openerPicker.setAttribute("data-s-picker", "s-bar-collapse");
        self.resizeBar.appendChild(self.openerPicker);
        self._bar.appendChild(self.resizeBar);
        self.openerPickerHeight = $(self.openerPicker).outerHeight();
        self._resizer = document.site.setResizable({
            source: self,
            slot: self.layoutSlot,
            dragSpot: null,
            direction: {
                right: true
            },
            minWidth: 180,
            resizerSpot: self.resizeBar,
            onResize: function(resizer, moving){
                if (!moving) {
                    document.site.toggleClass(self.resizeBar, "s-resizing", document.site.resizing);
                    document.site.toggleClass(self.openerPicker, "s-resizing", document.site.resizing);
                }
            }
        });
        self._resizer.isEnabled = !self.preferences.isCollapsed;
        
    },
    createBlock: function($item){
        var block = this.blocks[$item.$id] = {
            boxParent: this,
            $item: $item
        };
        var blockItem = document.createElement("section");
        blockItem.className = "s-aw-toolbar-block";
        blockItem.setAttribute("data-s-block", $item.$id);
        block.$$item = $(blockItem);
        
        var title = document.createElement("a");
        title.setAttribute("data-s-picker", "s-bar-title");
        title.className = "s-aw-toolbar-block-title";
        title.textContent = $item.$title || "";
        
        var dom = document.createElement("div");
        dom.className = "s-aw-toolbar-block-body";
        block.$$body = $(dom);
        
        dom = document.createElement("header");
        dom.className = "s-aw-toolbar-block-header";
        dom.appendChild(title);
        block.$$header = $(dom);
        
        block.$$item.append(block.$$header).append(block.$$body);
        this._body.appendChild(blockItem);
        
        block.$item.$isTitleHidden = true;
        block.$$body[0].style.display = "none";
        block.$$header.toggleClass("s-open", false);
        block.$$item.toggleClass("s-open", false);
        block.content = this.loadNewItem(block.$$body.empty()[0], block.$item);
    },
    collapseBlock: function(block, show, isFirstTime, onDrawBar){
        if (show && this.openedBlock && (this.openedBlock != block)) {
            this.collapseBlock(this.openedBlock, false);
        }
        this.openedBlock = block;
        block.$$body[0].style.display = show ? "" : "none";
        block.$$header.toggleClass("s-open", show);
        block.$$item.toggleClass("s-open", show);
        if (show) {
            this.awContext.treeContent.toggle(block.$item.$id == "$treeContentBlock");
            this.onWindowResize();
        }
        else {
            block.$$item[0].style.height = "";
        }
    },
    ensureState: function(){
        if (this.preferences.isCollapsed) {
            this.openedWidth = this.$$slot.outerWidth();
            this.layoutSlot.style.width = "10px";
            this._body.style.display = "none";
            this.resizeBar.className = "s-aw-toolbar-resizer s-close";
            this.openerPicker.className = "s-aw-toolbar-opener s-close";
            if (!this.preferences.isDocked) {
                this._bar.style.width = "10px";
            }
        }
        else {
            this.resizeBar.className = "s-aw-toolbar-resizer";
            this.openerPicker.className = "s-aw-toolbar-opener";
            this.resizeBar.style.width = "";
            if (this.preferences.isDocked) {
                this.layoutSlot.style.width = (this.openedWidth || 180) + "px";
                this._bar.style.position = "";
                this._bar.style.width = "";
            }
            else {
                this.layoutSlot.style.width = "10px";
                document.site.setZIndex(this._bar);
                this._bar.style.position = "absolute";
                this._bar.style.width = "180px";
                this._bar.style.top = "0px";
                this._bar.style.left = "0px";
            }
            this._body.style.display = "";
        }
    },
    onWindowResize: function(){
        if (this.$$bar && this.layoutSlot) {
            var height = $(this.awContext.targetPage.scrollview).height();
            if (height) {
                this._bar.style.height = height + "px";
                this.resizeBar.style.height = height + "px";
                this.openerPicker.style.top = ((height - this.openerPickerHeight) / 2) + "px";
                this._body.style.height = (height - this._body.offsetTop) + "px";
                if (this.openedBlock) {
                    var blockHeight = 0;
                    var ids = Object.keys(this.blocks);
                    for (var ii = 0, jj = ids.length; ii < jj; ii++) {
                        var block = this.blocks[ids[ii]];
                        if (block != this.openedBlock) {
                            blockHeight += block.$$item.outerHeight(true);
                        }
                    }
                    blockHeight = this._body.clientHeight - blockHeight;
                    this.openedBlock.$$item[0].style.height = blockHeight + "px";
                    this.openedBlock.$$body.height(this.openedBlock.$$item.height() - this.openedBlock.$$header.outerHeight(true));
                    if (this.openedBlock.content && this.openedBlock.content.resize) {
                        this.openedBlock.content.resize(this.openedBlock.$$body.height());
                    }
                    if (this.openedBlock.onResize) {
                        this.openedBlock.onResize(this.openedBlock.$$body.height());
                    }
                }
            }
        }
    },
    applyChange: function(newData){
        if (newData) {
            if (newData.$isHidden !== undefined) {
                var keys = ["$addBlock", "$addSection", "$addMenus"];
                for (var ii = 0, jj = keys.length; ii < jj; ii++) {
                    if (this.menuItems[keys[ii]]) {
                        this.menuItems[keys[ii]][0].setMenu({
                            $isDisabled: newData.$isHidden
                        });
                    }
                }
            }
        }
        RawPage.prototype.applyChange.call(this, newData);
    },
    onMenuClick: function(menuItem){
        switch (menuItem.$item.$bind) {
            case "$addSection":
            case "$addBlock":
                var newChild = this.awContext.authorPage.addNewItem(this.awContext.targetPage, this.awContext.awItem, menuItem.$item.$bind.replace("$add", "").toLowerCase());
                this.awContext.authorPage.endTargetPageUpdate(newChild, true);
                return false;
            case "$addMenus":
                var newChild = this.awContext.authorPage.addNewItem(this.awContext.targetPage, this.awContext.awItem, "block", null, undefined, undefined, "menus");
                this.awContext.authorPage.endTargetPageUpdate(newChild, true);
                return false;
            case "$defaultLayout":
            case "pattern-headerTabs":
                this.awContext.palette.onChangeSep(menuItem);
                return false;
            default:
                if (menuItem.$item.$fieldBind) {
                    var newChild = this.awContext.drawHelper.onAddNewField(this.awContext, this.awContext.awItem, menuItem.$item);
                    this.awContext.authorPage.endTargetPageUpdate(newChild, true);
                }
                else {
                    if (menuItem.$item.$bind.indexOf("pattern") == 0) {
                        this.awContext.palette.onChangeSep(menuItem);
                        break;
                    }
                    this.awContext.awItem.applyDesignMetaData(metaData, true);
                }
                break;
        }
    },
    onExcludeField: function($bind, isExcluded){
        var $actions = {};
        $actions["$fieldLink" + $bind] = {
            $isHidden: isExcluded
        };
        this.applyChange({
            $actions: $actions
        });
    },
    toggleItemAuthoring: function(enable){
    }
});
