// etna-engine/lib/runtime/errors.js
"use strict";
var config = require('../../../../nodelocal').config.etna;
var trace  = (config.traceError)?console.log:null;

var _details = {
	1: ["ERPAFPA", "Il manque une parenthèse fermante"],
	2: ["ERPAOPA", "Il manque une parenthèse ouvrante"],
	3: ["ERPAVIR", "Il manque une virgule"],
	4: ["ERPAFON", "Fonction inconnue"],
	5: ["ERCARIN", "Caractère illégal"],
	6: ["ERVARI", "Variable inexistante"],
	7: ["ERCLAS", "Classe inexistante"],
	8: ["ERINDI", "Indice incorrect"],
	9: ["ERPAZC", "Pas de zone courante"],
	10: ["ERMODE", "Incompatibilité de type"],
	11: ["ERSQNG", "Racine d'un nombre négatif "],
	12: ["ERTRIG", "Erreur en calcul trigonométrique"],
	13: ["ERHUGE", "Dépassement de capacité en calcul"],
	14: ["ERLOGN", "Logarithme d'un nombre négatif"],
	15: ["ERTRIH", "Erreur trigonométrie hyperbolique"],
	16: ["ERFACT", "Erreur en calcul de factorielle"],
	17: ["ERINEG", "Indice négatif"],
	18: ["ERNULI", "Numéro de libellé incorrect"],
	19: ["INCFIC", "Type de fichier incorrect"],
	20: ["PAFIC", "Fichier inexistant"],
	21: ["PACLE", "Clé inexistante dans cette table"],
	22: ["MODIN", "Mode de lecture incorrect"],
	23: ["ERISAM", "Erreur en gestion de fichier C-ISAM"],
	24: ["ERGFIC", "Erreur gestion fichier séquentiel"],
	25: ["ERSYST", "Erreur système"],
	26: ["ERINT", "Erreur interne"],
	27: ["ERACCE", "Problème d'accès sur un fichier"],
	28: ["ERREOP", "Ouverture d'une même table 2 fois"],
	29: ["TROFIC", "Trop de tables ouvertes"],
	30: ["PAFIAS", "Mauvaise table associée au masque"],
	31: ["PLUMEM", "Plus de mémoire disponible"],
	32: ["ERRET", "Return ne correspondant pas à un gosub"],
	33: ["EREX", "Fonction non implémentée !"],
	34: ["ERVARC", "Erreur en création de variable"],
	35: ["ERPATB", "Le rang n'est pas un tableau"],
	36: ["NOFEN", "Fenêtre inexistante"],
	37: ["ERMABR", "2 masques de même abréviation"],
	38: ["TROPRO", "Trop de traitements ou de procédures en cours"],
	39: ["ERLAB", "Etiquette inexistante"],
	40: ["ERECR", "Erreur en écriture de fichier"],
	41: ["ERLOOP", "Valeur de step nulle"],
	42: ["ERTROM", "Trop de Masques Ouverts"],
	43: ["FISLOCK", "Table  verrouillée"],
	44: ["ERMDISK", "Plus de place sur le disque"],
	45: ["ERAPL", "Erreur en gestion du fichier APL"],
	46: ["ERIMP", "Fichier imprimante inexistant"],
	47: ["ERPAMK", "Masque Inexistant"],
	48: ["PADTRAN", "Pas de transaction en cours"],
	49: ["DEJTRAN", "Il y a déjà une transaction en cours"],
	50: ["ERDOM", "Fonction indéfinie pour la valeur donnée"],
	51: ["ERDCB", "Erreur en gestion des DCB"],
	52: ["ERLOGZ", "Erreur log(0)"],
	53: ["ERDIVZ", "Erreur division par 0"],
	54: ["ERSRAP", "Sérialisation de l'application incorrecte !"],
	55: ["ERDIM", "Trop de dimensions"],
	56: ["ERDATE", "Erreur date incorrecte"],
	57: ["ERSEEK", "Pas de seek possible sur stdio"],
	58: ["ERMESS", "Erreur sur le fichier message"],
	59: ["ETRETA", "Trop de Tableaux Ouverts"],
	60: ["ERCHAN", "Plus de canaux disponibles"],
	61: ["ERVEX", "Variable déjà  existante"],
	62: ["EREADONLY", "Variable non modifiable"],
	63: ["ACEXIMP", "Ouverture accès exclusif impossible"],
	64: ["NOPROC", "Processus inexistant"],
	65: ["ERULIM", "Dépassement limite taille fichiers"],
	66: ["ERFNAM", "Nom de fichier incorrect"],
	67: ["ERTRAD", "Trop d'Adonix Tournent"],
	68: ["ERTRAP", "Trop d'Utilisateurs utilisent cette application"],
	69: ["ERARGNO", "Mauvais nombre de paramètres"],
	70: ["ERARGTY", "Type de paramètre incompatible"],
	71: ["CLOTRAN", "Fermeture impossible durant une transaction"],
	72: ["ERADNS", "Adonix n'a pas été sérialisé"],
	73: ["ERAPIN", "Application inexistante"],
	74: ["ERAPAD", "Sérialisation ADONIX/Application incompatible"],
	75: ["ERORA", "Erreur Oracle"],
	76: ["ERING", "Erreur SQL Server"],
	77: ["ERRTOM", "Trop d'Options dans ce menu"],
	78: ["ERTRAN", "Erreur sur fichier de transaction"],
	79: ["ERTRWD", "Trop de fenêtres ouvertes"],
	80: ["ETERIN", "Erreur en initialisation du terminal"],
	81: ["ENAREA", "Zone inexistante"],
	82: ["ERTROFORM", "Trop d'états ouverts"],
	83: ["EROUVFORM", "Ouverture d'un même état 2 fois"],
	84: ["EROBEX", "Objet déjà existant"],
	85: ["", "Trop de slots"],
	86: ["", "Slot inexistant"],
	87: ["", "Slot déjà existant"],
	88: ["", "Connexion non authorisée"],
	89: ["", "Erreur DB2"],
	90: ["", "Erreur VBScript"],
	91: ["", "Trop de liens"],
	92: ["", "Duplication de clé sur un index unique"],
	93: ["", "Erreur LDAP"],
	94: ["", "Erreur d'authentification Kerberos"],
	95: ["", "Délai d'authentification expiré"],
	96: ["", "Connexion non autorisée"],
	97: ["", "La délégation n'est pas autorisée pour ce service"],
	98: ["", "Difference horaire trop importante"],
	99: ["", "Erreur sur l'archive : Ecrasement du FIRST_HDA"],
	100: ["", "Erreur chainage du premier libre : Precedent different de 4"],
	101: ["", "Erreur lors du LOCK de l'archive pour modification"],
	102: ["", "Erreur lors du UNLOCK de l'archive"],
	103: ["", "Erreur lors du LOCK de l'index de l'archive"],
	104: ["", "Erreur lors du UNLOCK de l'index de l'archive"],
	105: ["", "Trop de constantes"],
	106: ["", "Table ou vue inaccessible"],
	107: ["", "Fichier inexistant dans l'archive"],
	108: ["", "Erreur le nom de fichier est supérieur à 12 caractères"],
	109: ["", "Utilisateur ou rôle inexistant"],
	110: ["", "La variable doit être une structure"],
	111: ["", "Opération impossible sur un champ de structure"],
	112: ["", "Membre de structure inexistant"],
	113: ["", "L'étiquette n'est pas une sous-procédure"],
	114: ["", "L'étiquette n'est pas une méthode"],
	115: ["", "Aucun objet courant"],
	116: ["", "ERREUR_XML : Document XML incorrect"],
	117: ["", "ERREUR_XML : Structure XML non parsable"],
	118: ["", "ERREUR_XML : Abréviation inconnue"],
	119: ["", "ERREUR_XLM : Abréviation déjà utilisée"],
	120: ["", "ERREUR_XML : Trop de fichiers XML ouverts"],
	121: ["", "ERREUR_XML : Noeud avec trop d'attributs"],
	122: ["", "ERREUR_XML : Pas de nom de fichier"],
	123: ["", "ERREUR_XML : Il manque l'abréviation"],
	124: ["", "ERREUR_XML : Il manque le nom du noeud"],
	125: ["", "ERREUR_XML : Il n'existe pas de noeud avec ce nom"],
	126: ["", "Canal en cours d'utilisation"],
	127: ["", "Erreur de conversion UTF8"],
	128: ["", "Nombre d'instructions ou d'expressions trop important"],
	129: ["", "Erreurs emises par le serveur distant"],
	130: ["", "Erreur de communication"],
	131: ["", "Canal de communication non ouvert"],
	200: ["", "Compilation error"],
};

var errmes$ = exports.errmes$ = function(errn) {
		var det = _details[errn];
		return det && ("[" + det[0] + "] " + det[1]);
	};

exports.moduleName = function(loc) {
	if (!loc.file) return "unknown";
	var file = loc.file;
	var s = file.substring( Math.max(file.lastIndexOf('/'),file.lastIndexOf('\\')) + 1);
	//console.log("moduleName: " + s);
	var dot = s.lastIndexOf('.');
	return /*"x3://host:port/folder/" +*/ dot >= 0 ? s.substring(0, dot) : s;
};

function X3Frame(file, line, name) {
	this.file = file, this.line = line, this.name = name;
}
X3Frame.prototype.toString = function() {
	return "  at " + (this.name || '<script>') + " (" + this.file + (this.line ? ":" + this.line : "") + ")";
};

function X3Error(errn, message, jsStack) {
	//Error.captureStackTrace(this, X3Error);
	function cut(s) { return s.substring(s.indexOf('\n') + 1);}
	if (arguments.length === 1) {
		// errn is an Error object
		message = errn.message;
		jsStack = cut(errn.stack);
		errn = 20;
	}
	this.message = "[" + errn + "] " + (message || errmes$(errn));
	this.errn = errn;
	if (arguments.length > 3) {
		this.frames = [new X3Frame(arguments[2], arguments[3], arguments[4])];
		this.jsStack = cut(new Error('').stack);
	} else {
		this.frames = [];
		this.jsStack = jsStack || cut(new Error().stack);
	}
	if(trace) {
		trace("\nX3Error("+errn+"):"+message);
		var root;
		this.jsStack.split(")\n").some(function(line) {
			if(!root) {
				var node_modules = "node_modules";
				root = line.substring(line.indexOf("(")+1,line.indexOf(node_modules)+node_modules.length);
			}
			trace(line.replace(root,"."));
			return line.indexOf("etna-engine") >= 0;

		});
	}
}

X3Error.prototype.fill = function(frame) {
	for (; frame; frame = frame.prev) {
		if (frame.fcall) this.frames.push(new X3Frame("<built-in functions>", 0, frame.fcall));
		for (var sub = frame.sub; sub; sub = sub.prevSub) {
			this.frames.push(new X3Frame(sub.loc.file, sub.loc.line, sub.name || frame.prog.name));
		}
	}
	this.fill = null; // fill only once	
};
X3Error.prototype.toString = function() {
	return this.message;
};
Object.defineProperty(X3Error.prototype, 'stack', {
	get: function() {
		//console.log(this);
		return this.message + //
		"\n  <<< X3 stack >>>\n" + this.frames.join("\n") + //
		"\n  <<< JS stack >>>\n" + this.jsStack;	
	}
});

exports.X3Error = X3Error;
