// etna-engine/lib/runtime/numbers.js
"use strict";
var util = require('./util');
var errors = require('./errors');
var X3Error = errors.X3Error;
var glob = require('streamline/lib/globals');
var bcd = require('etna-engine/lib/runtime/bcd') ;
var operators = require('./operators').binaryOps;
var integer = require('./integer')

function fac(n) {
	return (n > 1) ? n * fac(n - 1) : 1;
}
function anp(n, p) {
	return fac(n)/fac(n-p);
}

function toRadians(angle) {
	var frame = glob.context.x3frame;
	switch(frame.context.sys.values.adxmda) {
		case 0 : return angle * (Math.PI / 180); //Degrees => Radians
		case 2 : return angle * (Math.PI / 200); //Grades => Radians
	}
	return angle;
}

exports.functions = {
	ABS: {
		count: 1,
		fn: function(x) {
			util.checkNumber(x);
			return Math.abs(x);
		}
	},
	MOD: {
		count: 2,
		fn: function mod(x, y) {
			util.checkNumber(x);
			util.checkNumber(y);
			util.checkNotZero(y);
			return x >= 0 ? x % y : -(-x % y);
		}
	},
	RND: {
		count: 1,
		fn: function(x) {
			return x * Math.random();
		}
	},
	SGN: {
		count: 1,
		fn: function(x) {
			return x == 0 ? 0 : x > 0 ? 1 : -1;
		}
	},
	SQR: {
		count: 1,
		fn: function(v) { 
			if (typeof v != "number") {
				if (bcd.isBCD(v)) v = bcd.toDouble( v ) ;
				else if (integer.isInteger(v)) v = integer.toDouble( v ) ;
				else util.badOperand(v) ;
			}

			return  parseFloat(Math.sqrt(v).toFixed(10)) ;
		}
	},
	AR2: {
		count: 1,
		fn: function(v) {
			return Math.floor(v * 100 + 0.5) / 100;
		}
	},
	ARR: {
		count: 2,
		fn: function(x, y) {
			if ((typeof y === "number") && (y === 0)) return( x ) ;
			if (bcd.isBCD(y) && bcd.isZero(y)) return ( x ) ;
			if (integer.isInteger(y) && integer.isZero(y)) return x ;
						
			var val = operators.ADD( operators.DIV( x, y ) , bcd.fromDouble(0.5) ) ;			
			if (bcd.isBCD(val)) val = bcd.fromInt32(bcd.toInt32(val)) ;
			else if (typeof val === number) val = Math.floor(val) ;

			return operators.MUL( y, val ) ;
		}
	},
	FIX: {
		count: 1,
		fn: function(v) {
			return v >= 0 ? Math.floor(v) : -Math.floor(-v);
		}
	},
	INT: {
		count: 1,
		fn: Math.floor

	},
	ANP: {
		count: 2,
		fn: anp
	},
	CNP: {
		count: 2,
		fn: function(n, p) {
			return anp(n, p) / anp(p);
		}
	},
	FAC: {
		count: 1,
		fn: fac
	},
	COS: {
		count: 1,
		fn: function(angle) {
			if (typeof angle != "number") {
				if (bcd.isBCD(angle)) angle = bcd.toDouble( angle ) ;
				else if (integer.isInteger(angle)) angle = integer.toDouble( angle ) ;
				else util.badOperand(angle) ;
			}	

			return parseFloat(Math.cos(toRadians(angle)).toFixed(10)) ;
		}
	},
	SIN: {
		count: 1,
		fn: function(angle) {
			if (typeof angle != "number") {
				if (bcd.isBCD(angle)) angle = bcd.toDouble( angle ) ;
				else if (integer.isInteger(angle)) angle = integer.toDouble( angle ) ;
				else util.badOperand(angle) ;
			}	

			//return parseFloat(Math.sin(toRadians(angle)).toFixed(10)) ;
			return Math.sin(toRadians(angle)) ;
		}
	},
	TAN: {
		count: 1,
		fn: function(angle) {
			return Math.tan(toRadians(angle));
		}
	},
	ACOS: {
		count: 1,
		fn: Math.acos
	},
	ASIN: {
		count: 1,
		fn: Math.asin
	},
	ATAN: {
		count: 1,
		fn: Math.atan
	},
	ACH: {
		count: 1,
		fn: function(x) {
			return Math.log(x + Math.sqrt(x * x - 1));
		}
	},
	ASH: {
		count: 1,
		fn: function(x) {
			return Math.log(x + Math.sqrt(x * x + 1));
		}
	},
	ATH: {
		count: 1,
		fn: function(x) {
			return Math.log((1 + x) / (1 - x)) / 2;
		}
	},
	CH: {
		count: 1,
		fn: function(x) {
			return (Math.exp(x) + Math.exp(-x)) / 2;
		}
	},
	SH: {
		count: 1,
		fn: function(x) {
			return (Math.exp(x) - Math.exp(-x)) / 2;
		}
	},
	TH: {
		count: 1,
		fn: function(x) {
			return (Math.exp(2 * x) - 1) / (Math.exp(2 * x) + 1);
		}
	},
	EXP: {
		count: 1,
		fn: Math.exp
	},
	LN: {
		count: 1,
		fn: Math.log
	},
	LOG: {
		count: 1,
		fn: function(val) {
			if (typeof val === "number") return parseFloat( (Math.log(val) / Math.LN10).toFixed(10) );
			else if (integer.isInteger(val)) return  parseFloat( (Math.log(integer.toDouble(val)) / Math.LN10).toFixed(10) ) ;
			else if (bcd.isBCD(val)) return parseFloat( (Math.log(bcd.toDouble(val)) / Math.LN10).toFixed(10) ) ;
			else throw util.badOperand(val) ;
		}
	},
	PI: {
		count: 0,
		fn: function() {
			return Math.PI ;
		}
	},
	SETBIT: {
		count: 3,
		fn: function(val,pos,bit) {
			val |= (bit << (pos-1));
			return val;
		}
	},
	GETBIT: {
		count: 2,
		fn: function(val,pos) {
			return (val & (1 << (pos-1)))?1:0;
		}
	}
};
