// etna-engine/lib/runtime/tblbfile.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var util = require('./util');
var tdate = require('./tdate');
var tdatetime = require('./tdatetime');
var datetime = require('syracuse-core/lib/types/datetime');
var tbcd = require('./tbcd');
var Clbfile = require('./tclbfile').Clbfile;
var X3Error = require('./errors').X3Error;

class Blbfile {
	constructor(v) {
		var self = this;
		if (v instanceof Buffer) {
			self.value = v;
		} else throw util.badOperand(v);
	}
	valueOf() {
		var self = this;
		return self.value;
	}
	x3IsNumeric() {
		return false;
	}
	x3AssertOverflow(type) {
		return true;
	}
	x3Minus(v) {
		throw util.badOperand(v);
	}
	x3ToDouble(v) {
		throw util.badOperand(v);
	}
	x3ToString(t) {
		var self = this;
		if (t !== undefined) return this.value.toString(t === 50 ? 'binary' : 'utf8');
		else {
			var b0 = self.valueOf();
			var b1 = new Buffer(b0.length);
			var v, i = 0;

			// to BE and trailling 0
			while (v = b0.readUInt16BE(i)) {
				b1.writeUInt16LE(v, i);
				i += 2;
			}
			return b1.toString("ucs2", 0, i);

		}
	}
	x3ToBase64() {
		return this.value.toString('base64');
	}
	x3ToBCD(v) {
		return tbcd.fromBlbfile(this);
	}
	x3ToClbfile() {
		var self = this;
		var b0 = self.valueOf();
		var len = b0.readUInt32BE(0);
		var v, i = 4;
		if (len === 0) return "";
		var b1 = new Buffer(len);

		// to BE and trailling 0
		while (v = b0.readUInt16BE(i)) {
			b1.writeUInt16LE(v, i - 4);
			i += 2;
		}
		return new Clbfile(b1.toString("ucs2", 0, i - 4));
	}
	x3ToInteger() {
		var self = this;
		var b = this.value;
		var uint8 = function(v) {
			return (v < 0) ? v + 256 : v;
		};
		var int32 = function(v) {
			return (v >= Math.pow(2, 31)) ? v - Math.pow(2, 32) : v;
		};
		var int16 = function(v) {
			return (v >= Math.pow(2, 15)) ? v - Math.pow(2, 16) : v;
		};

		switch (b.length) {
			case 8:
				var hi = b.readUInt32BE(0);
				var lo = b.readUInt32BE(4);
				return hi * 0x100000000 + lo;
			case 4:
				return b.readUInt32BE(0);
			case 3:
				return int32((uint8(b.readInt8(0)) << 16) + (uint8(b.readInt8(1)) << 8) + uint8(b.readInt8(2)));
			case 2:
				return b.readInt16BE(0);
			case 1:
				return b.readUInt8(0);
			default:
				throw new X3Error(26, "Bad size conversion to Integer");
		}
	}
	x3ToDate() {
		var self = this;

		var o = tdate.ORIGIN();
		return (o.x3Add(this.x3ToInteger() - 1));
	}
	x3ToUuid() {
		var self = this;
		return self.value;
	}
	x3ToDatetime(v) {
		var n = this.x3ToInteger(8);
		var sec = n % 60;
		var m = Math.floor(n / 60);
		var min = m % 60;
		var h = Math.floor(m / 60);
		var ho = h % 24;
		var d = Math.floor(h / 24);
		var o = tdate.ORIGIN();
		var dt = o.x3Add(d - 1);
		return datetime.make(dt.year, dt.month, dt.day, ho, min, sec, 0);
	}
	x3Serialize(v) {
		return this.value;
	}
	x3IsZero() {
		throw util.badOperand(this);
	}
	x3IsEmpty() {
		return this.value.length === 0;
	}
	x3IsUndefined() {
		throw util.badOperand(this);
	}
	x3Length() {
		var self = this;
		return self.value.length;
	}
	x3Compare(b) {
		var self = this;
		if (!util.x3IsBlbfile(b))
			throw util.badOperand(b);

		if (self.value.length < b.value.length) return -1;
		else if (self.value.length > b.value.length) return 1;
		else {
			for (var i = 0; i < this.value.length; i++) {
				if (self.value[i] < b.value[i]) return -1;
				else if (self.value[i] > b.value[i]) return 1;
			}
			return 0;
		}
	}
	x3Add(b) {
		throw util.badOperand(b);
	}
	x3Sub(b) {
		throw util.badOperand(b);
	}
	x3Mul(b) {
		throw util.badOperand(b);
	}
	x3Div(b) {
		throw util.badOperand(b);
	}
	x3Mod(b) {
		throw util.badOperand(b);
	}
	x3Pow(b) {
		throw util.badOperand(b);
	}
	x3Xor(b) {
		throw util.badOperand(b);
	}
	x3Abs(v) {
		throw util.badOperand(v);
	}
	x3Truncate(v) {
		throw util.badOperand(v);
	}
	x3ToSql(driver) {
		var self = this;
		return driver.toSqlLob(this);
	}
	x3Clone() {
		return new Blbfile(this.value);
	}
	x3Type() {
		return 522;
	}
}

/*function fromString(s, t) {
	return new Blbfile(new Buffer(s, t === 50 ? 'binary' : 'utf8'));
}*/

function fromBase64(s) {
	return new Blbfile(new Buffer(s, 'base64'));
}

exports.Blbfile = Blbfile;

exports.functions = {
	TOBLBFILE: {
		count: 2,
		fn: function(c, t) {
			c = util.x3ToString(c, true);
			if (typeof t !== 'number') throw util.badOperand(t);
			return new Blbfile(new Buffer(c, t === 50 ? 'binary' : 'utf8'));
		}
	},
	TOCLBFILE: {
		count: 2,
		fn: function(b, t) {
			if (!util.x3IsBlbfile(b)) throw util.badOperand(b);
			if (typeof t !== 'number') throw util.badOperand(t);
			return new Clbfile(b.value.toString(t === 50 ? 'binary' : 'utf8'));
		}
	},
	/*
	TOBASE64: {
		count: 1,
		fn: function(b) {
			if (util.x3IsBlbfile(b)) return b.x3ToBase64();
			b = util.x3ToString(b, true);
			return fromString(b).x3ToBase64();
		}
	},
	FROMBASE64: {
		count: 1,
		fn: function(c) {
			c = util.x3ToString(c, true);
			return fromBase64(c);
		}
	},
	*/
	B64DECODE: {
		count: 2,
		fn: function(clb, blb) {
			var str = util.x3ToString(clb, true);
			if (!util.x3IsBlbfile(blb)) throw util.badOperand(blb);
			blb.value = new Buffer(str, 'base64');
			return blb.value.length;
		}
	},
	B64ENCODE: {
		count: 2,
		fn: function(blb, clb) {
			if (!util.x3IsBlbfile(blb)) throw util.badOperand(blb);
			if (!util.x3IsClbfile(clb)) throw util.badOperand(clb);
			clb.value = blb.value.toString('base64');
			return clb.value.length;
		}
	},
};

//exports.fromString = fromString;
//exports.fromBase64 = fromBase64;
