// etna-engine/lib/runtime/tdatetime.js
"use strict";
var X3Error = require('./errors').X3Error;
var util = require('./util');
var bcd = require('./Datetime');

exports.isDatetime = function(val) {
  return val instanceof Datetime ;
}

var Datetime = (function() {
  function Datetime( v ) {

    } ;

    Datetime.prototype.toDatetime = function() {
      return this ;
    } ;

    Datetime.prototype.toString = function( v ) {
      return ""+this.value ;
    } ;

    Datetime.prototype.toInteger = function() {
      return new Datetime( this ) ;
    } ;
  
    Datetime.prototype.toDouble = function() {
      return new Datetime( this ) ;
    } ;

    Datetime.prototype.toDate = function() {
      throw util.badOperand(this.value) ;
    } ;   

    Datetime.prototype.toBCD = function() {
      throw util.badOperand(this.value) ;
    } ;   

  Datetime.prototype.isNumeric = function() {
    return true ;
  } ;

    Datetime.prototype.isZero = function() {
    return (this.value === 0)?true : false ;
  } ; 

    Datetime.prototype.compare = function( b ) {

    } ; 

  Datetime.prototype.add = function( v ) {
    
    } ;

  Datetime.prototype.sub = function( v ) {
    
    } ;

  Datetime.prototype.mul = function( v ) {
    
    } ;

  Datetime.prototype.div = function( v ) {
    
    } ;

  Datetime.prototype.mod = function( v ) {
    
    } ;

  Datetime.prototype.pow = function( v ) {
    
    } ;

    Datetime.prototype.xor = function( v ) {
    
    } ;

    Datetime.prototype.abs = function( ) {
      return new Datetime( Math.abs( this.value )) ;
    } ;

    Datetime.prototype.truncate = function( v ) {
      return new Datetime( this.value ) ;
    } ;

})( ) ;

exports.Datetime = Datetime ;
