// etna-engine/lib/runtime/tdate.js
"use strict";
var X3Error = require('./errors').X3Error;
var util = require('./util');
var bcd = require('./Date');

exports.isDate = function(val) {
  return val instanceof Date ;
}

var Date = (function() {
  function Date( v ) {

    } ;

    Date.prototype.toDate = function() {
      return this ;
    } ;

    Date.prototype.toString = function( v ) {
      return ""+this.value ;
    } ;

    Date.prototype.toInteger = function() {
      return new Date( this ) ;
    } ;
  
    Date.prototype.toDouble = function() {
      return new Date( this ) ;
    } ;

    Date.prototype.toDatetime = function() {
      throw util.badOperand(this.value) ;
    } ;   

    Date.prototype.toBCD = function() {
      throw util.badOperand(this.value) ;
    } ;   

  Date.prototype.isNumeric = function() {
    return true ;
  } ;

    Date.prototype.isZero = function() {
    return (this.value === 0)?true : false ;
  } ; 

    Date.prototype.compare = function( b ) {

    } ; 

  Date.prototype.add = function( v ) {
    
    } ;

  Date.prototype.sub = function( v ) {
    
    } ;

  Date.prototype.mul = function( v ) {
    
    } ;

  Date.prototype.div = function( v ) {
    
    } ;

  Date.prototype.mod = function( v ) {
    
    } ;

  Date.prototype.pow = function( v ) {
    
    } ;

    Date.prototype.xor = function( v ) {
    
    } ;

    Date.prototype.abs = function( ) {
      return new Date( Math.abs( this.value )) ;
    } ;

    Date.prototype.truncate = function( v ) {
      return new Date( this.value ) ;
    } ;

})( ) ;

exports.Date = Date ;
