// etna-engine/lib/runtime/tinteger.js
"use strict";
var X3Error = require('./errors').X3Error;
var util = require('./util');
var bcd = require('./bcd');


exports.isInteger = function(val) {
	return val instanceof Integer ;
}

var Integer = (function() {
	function Integer( v ) {

    } ;

  	Integer.prototype.toInteger = function() {
    	return this ;
  	} ;

  	Integer.prototype.toString = function( v ) {
    	return ""+this.value ;
  	} ;

   	Integer.prototype.toDouble = function() {
    	return new Double( this ) ;
  	} ;
	
   	Integer.prototype.toBCD = function() {
    	return new BCD( this ) ;
  	} ;

	Integer.prototype.toDate = function() {
    	throw util.badOperand(this.value) ;
  	} ; 	

  Integer.prototype.toDatetime = function() {
      throw util.badOperand(this.value) ;
    } ;       

	Integer.prototype.isNumeric = function() {
		return true ;
	} ;

  	Integer.prototype.isZero = function() {
		return (this.value === 0)?true : false ;
	} ;	

  	Integer.prototype.compare = function( b ) {

  	} ;	

	Integer.prototype.add = function( v ) {
    
  	} ;

	Integer.prototype.sub = function( v ) {
    
  	} ;

	Integer.prototype.mul = function( v ) {
    
  	} ;

	Integer.prototype.div = function( v ) {
    
  	} ;

	Integer.prototype.mod = function( v ) {
    
  	} ;

	Integer.prototype.pow = function( v ) {
    
  	} ;

  	Integer.prototype.xor = function( v ) {
    
  	} ;

  	Integer.prototype.abs = function( ) {
    	return new Integer( Math.abs( this.value )) ;
  	} ;

  	Integer.prototype.truncate = function( v ) {
  		return new Integer( this.value ) ;
  	} ;

})( ) ;

exports.Integer = Integer ;
