// etna-engine/lib/runtime/tstring.js
"use strict";
var X3Error = require('./errors').X3Error;
var util = require('./util');
var bcd = require('./bcd');

exports.isString = function(val) {
	return val instanceof String ;
}

  var String = (function() {
  	function String( v ) {

      } ;

    	String.prototype.toString = function() {
      	return this ;
    	} ;

     	String.prototype.toDouble = function() {
      	return new Double( this ) ;
    	} ;
  	
     	String.prototype.toBCD = function() {
      	return new BCD( this ) ;
    	} ;

      String.prototype.toInteger = function() {
        
      } ;    

  	   String.prototype.toDate = function() {
        
    	} ; 	

       String.prototype.toDatetime = function() {
        
      } ;   

      String.prototype.isNumeric = function() {
  		  return false ;
      } ;

    	String.prototype.isZero = function() {
  		 
  	  } ;	

    	String.prototype.compare = function( b ) {

    	} ;	

  	String.prototype.add = function( v ) {
      
    	} ;

  	String.prototype.sub = function( v ) {
      
    	} ;

  	String.prototype.mul = function( v ) {
      
    	} ;

  	String.prototype.div = function( v ) {
      
    	} ;

  	String.prototype.mod = function( v ) {
      
    	} ;

  	String.prototype.pow = function( v ) {
      
    	} ;

    	String.prototype.xor = function( v ) {
      
    	} ;

    	String.prototype.abs = function( ) {
      	
    	} ;

    	String.prototype.truncate = function( v ) {
    		
    	} ;

  })( ) ;

  exports.String = String ;
