// etna-engine/lib/runtime/tstring.js
"use strict";
var X3Error = require('./errors').X3Error;
var util = require('./util');
var bcd = require('./bcd');
var tinteger = require('./tinteger');
var tdate = require('./tdate');

exports.isString = function(val) {
	return val instanceof String ;
}

var String = (function() {
	function String( v ) {
		this.value=v;
	} ;

	String.prototype.toString = function() {
		return this ;
	} ;
	
	String.prototype.toDouble = function() {
		return new Double( this ) ;
	} ;
	
	String.prototype.toBCD = function() {
		return new BCD( this ) ;
	} ;
	
	String.prototype.toInteger = function() { 
		return new tinteger.Integer(parseInt(this.value) ) ;
	} ;    
	
	String.prototype.toDate = function() {
		try {
			var y = this..substring(0,4);
			var m = this.substring(5,6);
			var d = this.substring(7,8);
			return new tdate.date(d,m,y) ;	
		} catch (e) {new tdate.date ; } ;
	
	} ; 	
	
	String.prototype.toDatetime = function() {
	        
	} ;   
	
	String.prototype.isNumeric = function() {
	return false ;
	} ;
	
	String.prototype.isZero = function() {
		if (this.value="") {
			return true;
		}else {
			return false;
		}
	} ;	
	
	String.prototype.compare = function( b ) {
		if (!exports.isString(b) throw util.badOperand(b);
		
		if (this.value === b.valueOf()) return 0 ;
		else if (this.value < b.valueOf()) return -1 ;
		else return 1 ;
	}	
	
	String.prototype.add = function( b ) {
		if (!exports.isString(b)){
			if (exports.isDate(b)) return this.value+b.valueof.toString("yyyyMMdd") ;
			else throw util.badOperand(b);	
		}
		return this.value + b.valueOf( ) ;
	} ;
	
	String.prototype.sub = function( v ) {
	} ;
	
	String.prototype.mul = function( v ) {
	} ;
	
	String.prototype.div = function( v ) {
	} ;
	
	String.prototype.mod = function( v ) {
	} ;
	
	String.prototype.pow = function( v ) {
	} ;
	
	String.prototype.xor = function( v ) {
	} ;
	
	String.prototype.abs = function( ) {
	} ;
	
	String.prototype.truncate = function( v ) {
	} ;

})( ) ;


exports.String = String ;
