// syracuse-ui/lib/field/array/selector/buttonRecordSelector.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var RecordSelector = require('./recordSelector').RecordSelector;

function ButtonRecordSelector() {}

exports.ButtonRecordSelector = helpers.defineClass(ButtonRecordSelector, RecordSelector, {
	load: function(list) {
		this.useColumnSelector = true;
		this.text = list.localize.flSelect || "Select";
		this.width = (11 * this.text.length);
		RecordSelector.prototype.load.call(this, list);
	},
	getWidth: function() {
		return this.width;
	},
	appendToTitle: function(slot) {

	},
	appendToRecord: function(slot) {
		slot.className += " s-list-selector-button-slot";
		var btn = document.createElement("a");
		btn.setAttribute("data-s-picker", "list-selector");
		btn.className = "s-list-selector-button";
		btn.textContent = this.text;
		btn.title = this.text;
		slot.appendChild(btn);
		return btn;
	},
	onclick: function(options) {
		options.isSelected = true;
		options.uuidTarget = this.list.findRecordUiid(options.selectorTarget);
	}
});
