// syracuse-ui/lib/field/array/selector/recordSelector.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

var _types = {
	none: {
		load: function(selector) {

		}
	},
	multi: {
		load: function(selector) {
			selector.isMulti = true;
			selector.useColumnSelector = !selector.list.$item.$selectByRowIndex && !selector.list.$prototype.$treeview;
		},
		appendToRecord: function(selector, slot) {
			return selector.addCheckBox(slot, true);
		}
	},
	single: {
		load: function(selector) {
			selector.useColumnSelector = !selector.list.$item.$selectByRowIndex && !selector.list.$prototype.$treeview;
		},
		appendToRecord: function(selector, slot) {
			return selector.addCheckBox(slot, false);
		}
	},
	button: {
		load: function(selector) {
			selector.useColumnSelector = true;
			selector.text = syra_local.flSelect || "Select";
		},
		appendToRecord: function(selector, slot) {
			slot.className += " s-list-selector-btn-slot";
			return slot.appendChild(syra_menus.addTextButton(selector.text, "s-list-selector-btn", "onSelectorClick"));
		}
	},
	row: {
		load: function(selector) {
			selector.isRowMode = true;
		}
	}
};

function _onMultiSelect(selector, targetRecord, unselectable) {
	var records = selector.list.records;
	var range = _getRange(selector, targetRecord);
	if (range) {
		if (range.targetIndex <= range.firstIndex) {
			for (var ii = range.targetIndex; ii < range.firstIndex; ii++) {
				selector.selectRecord(records[ii].$uuid, true);
			}
			for (ii = range.firstIndex + 1; ii <= range.lastIndex; ii++) {
				selector.selectRecord(records[ii].$uuid, false);
			}
		} else {
			for (var ii = range.firstIndex; ii <= range.targetIndex; ii++) {
				!records[ii].$isSelected && selector.selectRecord(records[ii].$uuid, true);
			}
			for (ii; ii <= range.lastIndex; ii++) {
				selector.selectRecord(records[ii].$uuid, false);
			}
		}
		if (targetRecord._selectorCheck && !targetRecord._selectorCheck.checked) {
			targetRecord._selectorCheck.checked = true;
		}
	}
}



function _onSelectAll(selector, isSelected, unselectable) {
	selector.records = {};
	var records = selector.list.records;
	for (var ii = 0, jj = records.length; ii < jj; ii++) {
		var $uuid = records[ii].$uuid;
		if (!(unselectable &&
			unselectable(selector.list, {
				"uuid": $uuid
			}))) {
			selector.selectRecord($uuid, isSelected);
		}
	}
}

function _getRange(selector, targetRecord) {
	var records = selector.list.records;
	var sels = selector.getSelectedList();
	if (sels.length) {
		return {
			firstIndex: records.indexOf(sels[0]),
			lastIndex: records.indexOf(sels[sels.length - 1]),
			targetIndex: records.indexOf(targetRecord)
		};
	}
	return null;
}

function _ensureContiguousSelection(selector, targetRecord) {
	var range = _getRange(selector, targetRecord);
	if (range) {
		var records = selector.list.records;
		if (range.targetIndex < (range.firstIndex - 1) || range.targetIndex > (range.lastIndex + 1)) {
			//out of range clear all
			for (var ii = range.firstIndex; ii <= range.lastIndex; ii++) {
				selector.selectRecord(records[ii].$uuid, false);
			}
		} else {
			if ((range.targetIndex > range.firstIndex) || (range.targetIndex < range.lastIndex)) {
				for (var ii = range.targetIndex + 1; ii <= range.lastIndex; ii++) {
					selector.selectRecord(records[ii].$uuid, false);
				}
			}
		}
	}
}

function _onSelectOn(selector, targetRecord, isSelected) {
	if (selector.isMulti && selector.$useContiguousSelection) {
		//_ensureContiguousSelection(selector, targetRecord);
		var range = _getRange(selector, targetRecord);
		if (range) {
			var records = selector.list.records;
			if (range.targetIndex < range.firstIndex || range.targetIndex > range.lastIndex) {
				//out of range clear all
				for (var ii = range.firstIndex; ii <= range.lastIndex; ii++) {
					selector.selectRecord(records[ii].$uuid, false);
				}
			} else {
				if ((range.targetIndex > range.firstIndex) || (range.targetIndex < range.lastIndex)) {
					for (var ii = range.targetIndex + 1; ii <= range.lastIndex; ii++) {
						selector.selectRecord(records[ii].$uuid, false);
					}
				}
			}
		}
	}
	selector.selectRecord(targetRecord.$uuid, isSelected !== false);
	if (selector.list.$item.$executeUrl && isSelected) {
		var uuids = Object.keys(selector.records);
		var $url = uuids && uuids.length && selector.records[uuids[0]].dataset.$url;
		if ($url) {
			syra_controller.executeMenu({
				$url: $url
			}, selector.records[uuids[0]]);
		}
	}
}

exports.onSelectorClick = function(list, event) {
	var selector = list.selector;
	var target = event.target;
	var record = selector.list.findRecord(target);
	var uuidTarget = target.syraSelectAll ? null : record.$uuid;
	var isSelected = target.checked !== false;
	if (selector.list.$item.$selectByRowIndex) {
		isSelected = selector.isMulti ? !selector.records[uuidTarget] : true; //byrowIndex => single always true else toggle
	}
	list.page.externalAdapter.onSelectRecordEvent({
		field: list,
		event: event,
		uuidTarget: uuidTarget,
		isSelected: isSelected,
		doEvent: function(unselectable) {
			if (!event.shiftKey && !selector.isMulti) {
				var sels = selector.getSelectedList();
				for (var ii = 0, jj = sels.length; ii < jj; ii++) {
					if (sels[ii] != record) {
						selector.selectRecord(sels[ii].$uuid, false);
					}
				}
			} else {
				document.getSelection().removeAllRanges();
			}
			if (target.syraSelectAll) {
				_onSelectAll(selector, isSelected, unselectable);
			} else {
				if (event.shiftKey) {
					_onMultiSelect(selector, record, unselectable);
				} else {
					_onSelectOn(selector, record, isSelected);
				}
			}
			if (list.articleParent.onSelectRecord) {
				list.articleParent.onSelectRecord(selector.records);
			} else {
				list.page.onSelectRecord && list.page.onSelectRecord(selector.records);
			}
			list.selectorCardRecord && list.selectorCardRecord.onSelectRecord(selector.records);
		}
	});
};


function RecordSelector() {}

exports.RecordSelector = helpers.defineClass(RecordSelector, null, {
	load: function(list) {
		this.type = _types[list.$item.$selectMode || "none"];
		this.list = list;
		this.$useContiguousSelection = this.list.$field.$useContiguousSelection || this.list.$item.$useContiguousSelection;
		this.records = {};
		this.type.load(this);
	},
	addCheckBox: function(slot, isCheck) {
		slot.className += " s-list-selector-slot";
		var input = document.createElement("input");
		input.setAttribute("type", isCheck ? "checkbox" : "radio");
		input.syraOnClick = "onSelectorClick";
		syra_site.setSpecificAttributes(input);
		input.className = "s-list-selector";
		return slot.appendChild(input);
	},
	getSelectedList: function() {
		var sels = [];
		var records = this.list.records;
		for (var ii = 0, jj = records.length; ii < jj; ii++) {
			records[ii].$isSelected && sels.push(records[ii]);
		}
		return sels;
	},
	appendToRecord: function(slot) {
		return this.type.appendToRecord(this, slot);
	},

	clear: function() {
		this.records = {};
		delete this.uuid;
	},
	unSelectAll: function() {
		var records = this.list.records;
		for (var ii = 0, jj = records.length; ii < jj; ii++) {
			if (records[ii].$isSelected) {
				this.selectRecord(records[ii].$uuid, false);
			}
		}
	},
	selectRecord: function(uuid, selected, selectFirstByDefault) {
		if (uuid === undefined) {
			if (Object.keys(this.records).length == 0) {
				if (this.list.records.length > 0) {
					this.selectRecord(this.list.records[0].$uuid, true);
				}
			}
		} else {
			if (!this.isMulti && selected) {
				if (this.uuid !== undefined && uuid != this.uuid) {
					this.selectRecord(this.uuid, false);
				}
			}
			var record = this.list.recordsMap[uuid];
			var dataset;
			if (record) {
				this.list.showRecordSelection(record, selected);
				dataset = record.ensureDataSet();
				record.$isSelected = selected;
			}
			if (selected && dataset) {
				this.uuid = uuid;
				this.records[uuid] = record;
				dataset.$isSelected = true;
			} else {
				delete this.uuid;
				delete this.records[uuid];
				if (dataset) {
					delete dataset.$isSelected;
				}
			}
			if (selectFirstByDefault && !record) {
				this.selectRecord(undefined, true);
				return;
			}
			if (this.list) {
				var page = this.list.page;
				if (selected && page && !page.isNavigationPage && page.lastSelectedList != this.list) {
					if (!this.list.$field.$isFusionNavigationList) {
						if (page.lastSelectedList && page.lastSelectedList.selector) {
							page.lastSelectedList.selector.unSelectAll();
						}
						this.list.page.lastSelectedList = this.list;
					}
				}
				this.list.selectorCardRecord && this.list.selectorCardRecord.onSelect(uuid, selected);
				this.list.treeDecorator && this.list.treeDecorator.onNodeSelected(record, selected);
			}
		}
	},
	onRecordApplyChange: function(record, $isSelected) {
		if (this.list) {
			if (record.isArticleLoaded) {
				_onSelectOn(this, record, $isSelected);
			} else {
				this.selectRecord(record.$uuid, $isSelected);
			}
		}
	},
	dispose: function() {
		this.list = this.records = null;
	}
});
