// syracuse-ui/lib/field/array/selector/selectorCardRecord.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Article = require("syracuse-ui/lib/article/article").Article;

function SelectorCardRecord() {}

exports.SelectorCardRecord = helpers.defineClass(SelectorCardRecord, Article, {
	dispose: function() {
		if (this._popup) {
			this._popup.dispose();
		}
		if (this._popupSlot) {
			document.site.removeDomChild(this._popupSlot);
		}
		this._item = this.slot = this._menusSlot = this._popup = this._popupSlot = null;
		Article.prototype.dispose.call(this);
	},
	initialize: function(list) {
		this.list = list;
		this.id = this.list.id + "-record-selector";
		this.list.$item.$isMenuRecordHidden = true;
		this.isRecordArticle = true;
		this.$defaultSkinBlock = "s-h3";
		this.$defaultSkinSection = "s-h3";
		this.arrayLevel = "record";
		this.$facet = "details";
		this.list.$isEditMode = this.$isEditMode = false;
	},
	loadBox: function() {
		//this.$uuid = options.dataRecord.$uuid;
		//record.$serverIndex = options.$serverIndex;
		this.$prototype = helpers.object.clone(this.list.$prototype.$item, true);
		this.$item = helpers.object.clone(this.list.$item.$selectorCard, true);
		if (!this.$item.$layout) {
			this.$item.$layout = {
				$items: this.list.$item.$layout.$items
			};
		}
		this._appendSelectorField();
		this.list.page.initializeNewItem(this, this.$item, this.list);
		Article.prototype.loadBox.call(this);
		this.applyChange(this.list.ensureGlobalMetaRecord());
	},
	_appendSelectorField: function() {
		if (!this.$item.$selectorField) {
			this.$item.$selectorField = Object.keys(this.$prototype.$properties)[0];
		}
		var $field = this.$prototype.$properties[this.$item.$selectorField];
		$field.$isHidden = true;
		this.$prototype.$properties.$selectorField = {
			$title: $field.$title,
			$type: "application/x-reference",
			$item: {
				$value: "{$selectField}",
				$key: "{$selectField}",
				$properties: {
					$selectField: {
						$type: $field.$type
					}
				},
				$selectField: "{" + this.$item.$selectorField + "}"
			}
		};
		this.placeSelectorField(this.$item.$layout);
	},
	placeSelectorField: function($layout) {
		if ($layout && $layout.$items) {
			for (var ii = 0, jj = $layout.$items.length; ii < jj; ii++) {
				var $item = $layout.$items[ii];
				if ($item.$bind == this.$item.$selectorField) {
					$layout.$items.splice(ii, 0, {
						$bind: "$selectorField",
						$isEditMode: true,
						$isTopLabelAlignment: false
					});
					break;
				} else {
					if ($item.$layout) {
						this.placeSelectorField($item.$layout);
					}
				}
			}
		}
	},
	drawBox: function() {
		this._item = document.createElement("div");
		this._item.setAttribute("data-s-record", this.$uuid);
		this._item.className = this.list.$skin + "-record-selector";
		this.$$item = $(this._item);
		var card = this._item.appendChild(document.createElement("div"));
		card.className = this.list.$skin + "-record";
		this.page.loadNewItem(card, this.$item, this);
		this.list.layoutSlot.appendChild(this._item);
		this._item.style.visibility = "";
	},
	onSelect: function($uuid, selected) {
		if (selected) {
			var found = this.list.findDataRecord($uuid);
			if (found) {
				this.applyChange({
					$properties: {
						$selectorField: {
							$links: this.$prototype.$links
						}
					}

				});
				this.applyChange(this.dataset = found.dataRecord);
			}
			if (this._popup) {
				this._popup.close();
				this._popup = null;
			}
		} else {
			this.dataset = null;
			var delta = {};
			var $binds = Object.keys(this.$prototype.$properties);
			for (var ii = 0, jj = $binds.length; ii < jj; ii++) {
				delta[$binds[ii]] = null;
			}
			this.applyChange(delta);
		}
	},
	onMenuClick: function(menuItem) {
		if (menuItem.$sourceBind && menuItem.$subRecordKey == "$selectorField") {
			document.controller.executeMenu(menuItem, null, this.list.recordsMap[this.dataset.$uuid]);
			return false;
		}
		return true;
	},
	onFieldClickPicker: function(field, pickerType, btn) {
		if ((field.$item.$bind == "$selectorField") && pickerType == "lookup") {
			var self = this;
			if (!self._popupSlot) {
				self._popupSlot = document.createElement("div");
				self._popupSlot.className = this.list.$skin + "-record-selector-popup";
			}
			self._popupSlot.appendChild(self.list._domItem);
			self._popup = self.openDialog({
				content: field,
				isPageBoundary: true,
				autoCloseBoundaryId: field.id,
				$dialogMode: "popup",
				$$dialog: $(self._popupSlot),
				position: {
					my: "left top",
					at: "left bottom",
					of: field.$$input
				},
				onClose: function() {
					self._popup = null;
				}
			});
			return false;
		}
		return true;
	},
	onNotifyDataChange: function(field, value) {
		if (field.$item.$bind == "$selectorField") {
			return false;
		}
		return true;
	},
	onWindowResize: function() {
		if (this._popup) {
			this._popup.onWindowResize(true);
		}
	}
});
