// syracuse-ui/lib/field/array/selector/singleRecordSelector.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var RecordSelector = require('./recordSelector').RecordSelector;

function SingleRecordSelector() {}

exports.SingleRecordSelector = helpers.defineClass(SingleRecordSelector, RecordSelector, {
	load: function(list) {
		this.useColumnSelector = !list.$item.$selectByRowIndex && !list.$prototype.$treeview;
		RecordSelector.prototype.load.call(this, list);
	},
	getWidth: function() {
		return 25;
	},
	appendToTitle: function(slot) {

	},
	appendToRecord: function(slot) {
		slot.className += " s-list-selector-slot";
		var input = document.createElement("input");
		input.setAttribute("type", "radio");
		input.setAttribute("data-s-picker", "list-selector");
		document.site.setSpecificAttributes(input);
		input.className = "s-list-selector";
		return slot.appendChild(input);
	},
	onclick: function(options) {
		options.isSelected = this.list.$item.$selectByRowIndex ? true : options.selectorTarget.checked;
		options.uuidTarget = this.list.findRecordUiid(options.selectorTarget);
	}
});
