// syracuse-ui/lib/jobs/jobsManager.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var DesktopPage = require("syracuse-ui/lib/page/desktopPage").DesktopPage;
var _formatApi = require('syracuse-ui/lib/field/formatApi');
var _date = require('syracuse-core/lib/types/date');
var _time = require('syracuse-core/lib/types/time');

var _severities = ["fatal", "error", "warning", "info", "success"];
var _$binds = {
	$links: ["$abort", "$suspend", "$resume", "$cancel", "$download", "$details"],
	$fields: ["title", "phase", "phaseDetail", "startTime", "startDate", "elapsedSeconds", "remainingSeconds"]
};

var _opener, _page, _popup;
var _jobs = {
	add: function(job) {
		if (!this.map) {
			this.map = {};
			this.list = [];
		}
		this.map[job.dataset.$uuid] = job;
		this.list.push(job);
		_opener.textContent = this.list.length;
		if (!_page) {
			_page = new JobsPage();
			_page.loadBox();
		}
		_opener.style.display = "";
		_page.toggle(true);
	},
	remove: function(job) {
		this.list.splice(this.list.indexOf(job), 1);
		_opener.textContent = this.list.length;
		if (this.list.length == 0) {
			_page.domItem.style.display = "none";
			_opener.style.display = "none";
		}
	}
};


function Job() {

}

helpers.defineClass(Job, null, {
	draw: function() {
		this.domItem = _page.scrollview.appendChild(document.createElement("div"));
		this.domItem.className = "s-jobs-job";

		var row = document.createElement("div");
		row.className = "s-job-top";
		this.typeIcon = row.appendChild(document.createElement("div"));

		var slot = document.createElement("div");
		slot.className = "s-job-progress";
		this.progressSlide = slot.appendChild(document.createElement("div"));
		this.progressSlide.className = "s-job-progress-bar";
		this.progress = slot.appendChild(document.createElement("div"));
		this.progress.className = "s-job-progress-data";

		this.jobBtn = syra_menus.addTextButton("", "s-job-action", "onJobActionClick", null, "cancel");
		this.jobBtn.style.visibility = "hidden";
		var clearBtn = syra_menus.addIconButton(syra_local.jobs_clear_job, "s-job-clear", "onClearJob", null, "delete");

		row.appendChild(slot);
		this.domItem.appendChild(row);

		var cell = document.createElement("div");
		cell.className = "s-job-action-cell";
		cell.appendChild(this.jobBtn);
		row.appendChild(cell);
		this.jobBtn.syraJob = clearBtn.syraJob = this.dataset.$uuid;
		this.jobBtn.syraItem = clearBtn.syraItem = _page.id;
		row.appendChild(clearBtn);

		this.cells = {};
		this.domItem.appendChild(row);

		for (var ii = 0, jj = _$binds.$fields.length; ii < jj; ii++) {
			var $bind = _$binds.$fields[ii];
			var slot = document.createElement("div");
			slot.style.display = "none";
			slot.className = "s-job-row";
			var title = document.createElement("div");
			title.textContent = syra_local["jobs_" + $bind];
			title.className = "s-job-title";
			var value = document.createElement("div");
			value.className = "s-job-value";
			this.cells[$bind] = {
				slot: this.domItem.appendChild(slot),
				title: slot.appendChild(title),
				value: slot.appendChild(value)
			};
		}
	},
	applyChange: function(newData, success, $url) {
		syra_site.deltaManager.applyObjectDelta(this, this.dataset, newData);
		this.dataset.$uuid = this.dataset.uuid;
		if (newData.$state != undefined) {
			this.dataset.phase = newData.$state;
		}
		this.dataset.$progressValue = success ? "success" : "on";
		if (newData.startDate) {
			var startDate = new Date(newData.startDate);
			this.dataset.startDate = _date.fromJsDate(startDate).toString();
			this.dataset.startTime = _time.fromJsDate(startDate).toString();
		}
		if (this.fusionContext) {
			// handling links in case of print job
			if (this.fusionContext.cancelSvcId && !success) {
				this.dataset.$links.$cancel = {
					$title: syra_local.jobs_cancel_label
				};
			} else {
				this.dataset.$links.$cancel = {
					$isHidden: true
				};
			}
			if (success && $url && $url.indexOf("/print/$report") != -1) {
				this.dataset.$links.$download = {
					$title: syra_local.jobs_download_label,
					$url: $url
				};
			}
		}
		if (newData.kind) {
			this.typeIcon.className = "s-job-kind-" + newData.kind;
		}
		if (newData.$links) {
			var visible;
			for (var mm = 0, pp = _$binds.$links.length; mm < pp; mm++) {
				var $bind = _$binds.$links[mm];
				var $link = newData.$links[$bind];
				if ($link && !$link.$isHidden) {
					var title = $link.$title || syra_local["jobs_" + $bind.replace("$", "") + "_label"];
					visible = true;
					syra_menus.updateButtonIcon(this.jobBtn, title, this.jobBtn.syraBind = $bind);
				}
			}
			this.jobBtn.style.visibility = visible ? "" : "hidden";
		}

		for (var ii = 0, jj = _$binds.$fields.length; ii < jj; ii++) {
			var $bind = _$binds.$fields[ii];
			var cell = this.cells[$bind];
			var show = "none";
			if (newData[$bind] != undefined) {
				show = "";
				switch ($bind) {
					case "startTime":
						cell.value.textContent = _page.timeApi.parse(newData.startTime).toString(_page.timeLocalFormat);
						break;
					case "startDate":
						cell.value.textContent = _page.datetApi.parse(newData.startDate).toString(_page.dateLocalFormat);
						break;
					default:
						cell.value.textContent = newData[$bind] || "";
						break;
				}
			}
			if (cell.slot.style.display != show) {
				cell.slot.style.display = show;
			}
		}

		newData.$diagnoses && this.fillDiagnoses(newData.$diagnoses);

		if (newData.$progressValue) {
			this.progressSlide.className = "s-job-progress-bar s-job-" + newData.$progressValue;
		}
		this.progress.textContent = newData.progress ? (newData.progress + " %") : "";

		_page.resizeArticle();
	},
	fillDiagnoses: function($diagnoses) {
		this.severityIndex = _severities.indexOf("success");
		if (!this.diagnoses.slot) {
			this.diagnoses.slot = this.diagnosesSlot = document.createElement("div");
			this.diagnoses.slot.className = "s-job-diagnoses";
			this.domItem.appendChild(this.diagnoses.slot);
		}
		for (var ii = this.diagnoses.items.length, jj = $diagnoses.length; ii < jj; ii++) {
			var $diagnose = $diagnoses[ii];
			var $severity = $diagnose.$severity || "error";
			var diagnose = {
				$diagnose: $diagnose,
				item: syra_menus.addFontIconText($diagnose.$message, "s-job-diagnose s-diag-" + $severity, $severity == "info" ? $severity : "diagnose")
			};
			if ($diagnose.$stackTrace) {
				this.diagnoses.hasStack = true;
				var stackBtn = diagnose.stackBtn = syra_menus.addIconButton(syra_local.jobs_stack_expand, "s-job-stack-opener", "onStackClick", null, "expand");
				stackBtn.$opened = false;
				stackBtn.syraJob = this.dataset.$uuid;
				stackBtn.syraDiag = ii;
				stackBtn.syraItem = _page.id;
				diagnose.item.insertBefore(stackBtn, diagnose.item.firstChild);
			}
			syra_site.dom.toggleClass(diagnose.item, "s-list-alt", ii % 2);
			this.diagnoses.slot.insertBefore(diagnose.item, this.diagnoses.slot.firstChild);
			this.diagnoses.items.push(diagnose);
			var severityIndex = _severities.indexOf($severity);
			if (severityIndex < this.severityIndex) {
				this.severityIndex = severityIndex;
			}
		}
		if (this.$progressValue == "success") {
			this.$progressValue = _severities[this.severityIndex];
		}
	},
	cancelAsynchRequest: function() {
		if (this.asynchReq && this.asynchReq.length > 0) {
			for (var ii = 0; ii < this.asynchReq.length; ii++) {
				clearTimeout(this.asynchReq[ii]);
			}
		}
	},
	callServer: function(reqOpt, cbck, onRelease) {
		var self = this;
		onRelease && self.cancelAsynchRequest();
		syra_controller.callServer(null, reqOpt, function(data, response, $url) {
			var reqOpt = {
				noDisplayErr: true,
				method: "GET",
				$location: {
					$url: response.headers.location
				}
			};
			if (self.fusionContext) {
				self.applyChange(data, response.status == 200, $url);
			} else {
				if (response.status != 204 && data != undefined) {
					// Cleaning diags records before filling them 
					self.applyChange({
						$diagnoses: []
					});
					self.applyChange(data, response.status == 200, $url);
				}
			}
			switch (response.status) {
				case 200:
					self.cancelAsynchRequest();
					if (self.fusionContext) {
						if ($url && $url.indexOf("/print/$report") != -1) {
							window.open($url, "_blank");
						}
						self.fusionContext && self.fusionContext.release();
					} else {
						// in case of asynch service operation, clean on success
						if ($url && $url.indexOf("/sdata/$trackers") != -1) {
							if (response.headers && response.headers['$do-not-delete']) {
								// Actually do not forward the final response to not erase async job data ($diagnoses...)
								// Do not DELETE tracker to be able to retrieve the final response throught $details link
								//self.callServer(this.buildReqOpt("GET", response.data.$links.$details.$url));
							} else {
								reqOpt.method = "DELETE";
								self.callServer(reqOpt);
							}
						}
					}
					break;
				case 201:
					if (self.fusionContext && response.headers.location.indexOf("/print/$report") != -1) {
						reqOpt.$acceptType = self.dataset.mime;
					}
					self.callServer(reqOpt);
					break;
				case 202:
					// 202 => server processes request asynchronously
					// do not send further request if a suspend request has been sent
					if ($url.indexOf("suspend=true") < 0) {
						self.asynchReq = self.asynchReq || [];
						self.asynchReq.push(setTimeout(function() {
							self.callServer(reqOpt);
						}, data.pollingMillis));
					}
					break;
				case 424:
					// received after tracker abort
					if ($url && $url.indexOf("/sdata/$trackers") != -1) {
						reqOpt.method = "DELETE";
						self.callServer(reqOpt);
					}
					break;
				default:
					//console.info("status=" + response.status);
					//console.log("data=" + data);
					break;
			}
		}, function(error, httpquery) {
			self.onRequestError(error);
		});
	},
	onRequestError: function(error, ignoreOpComplete) {
		var $diagnoses;
		if (error.data.indexOf("$diagnoses") != -1) {
			$diagnoses = JSON.parse(error.data).$diagnoses;
			if (!$diagnoses) {
				$diagnoses = [{
					$severity: "error",
					$message: syra_local.job_err_without_diag.replace("{status}", error.status)
				}];
			}
		} else {
			$diagnoses = [{
				$severity: "error",
				$message: error.data
			}];
		}
		this.applyChange({
			$diagnoses: $diagnoses
		}, true);
		syra_diagnose.closeBox("no", _page.msgBoxId);
		!ignoreOpComplete && this.fusionContext && this.fusionContext.release();
	},
	remove: function(checkKind) {
		_jobs.remove(this);
		syra_site.dom.removeChild(this.domItem);
		syra_site.disposeObject(this);
		(checkKind !== false) && _page.checkKindList();
	},
	onClientRelease: function($url, $method, cbck) {
		if (this.fusionContext) {
			if (this.fusionContext.cancelSvcId) {
				var $url = syra_site.expressionMaker.parse(this, this.dataset.$location, this.dataset.services[this.fusionContext.cancelSvcId]);
				this.callServer({
					noDisplayErr: true,
					method: "GET",
					$location: {
						$url: $url
					}
				}, cbck, true);
				this.fusionContext.release();
			}
		} else {
			$url &&
				this.callServer({
					noDisplayErr: true,
					method: $method,
					$location: {
						$url: $url
					}
				}, cbck, true);
		}
	}
});


function JobsPage() {

}

helpers.defineClass(JobsPage, DesktopPage, {
	onStackClick: function(event, btn) {
		btn.$opened = !btn.$opened;
		var state = btn.$opened ? "collapse" : "expand";
		syra_menus.updateButtonIcon(btn, syra_local["jobs_stack_" + state], state);
		var diagnose = _jobs.map[btn.syraJob].diagnoses.items[btn.syraDiag];
		if (btn.$opened && !diagnose.stackSlot) {
			diagnose.stackSlot = document.createElement("div");
			diagnose.stackSlot.className = "s-job-stack";
			diagnose.stackSlot.innerHTML = syra_site.dom.formatHTMLText(diagnose.$diagnose.$stackTrace);
			btn.parentNode.appendChild(diagnose.stackSlot);
		}
		syra_site.dom.display(diagnose.stackSlot, btn.$opened);
	},
	loadBox: function() {
		this.diagnosePage = syra_site;
		this.datetApi = _formatApi.getApi("application/x-date");
		this.dateLocalFormat = _formatApi.getLocalFormat("application/x-date");
		this.timeApi = _formatApi.getApi("application/x-time");
		this.timeLocalFormat = _formatApi.getLocalFormat("application/x-time");

		var $skin = "s-jobs";
		this.$prototype = {
			$properties: {
				kind: {
					$type: "application/x-choice",
					$value: {
						$type: "application/x-integer",
						$enum: []
					}
				}
			}
		};
		this.$item = {
			$layout: {
				$items: []
			}
		};

		this.layoutSlot = document.createElement("div");
		this.layoutSlot.className = $skin;
		syra_site.layoutSlot.appendChild(this.layoutSlot);

		DesktopPage.prototype.loadBox.call(this);
		this.header = document.createElement("div");
		this.header.className = "s-jobs-header";
		syra_site.dom.empty(this.domItem);
		this.domItem.appendChild(this.header);
		var div = document.createElement("div");
		div.className = "s-jobs-kind-slot";
		this.kindField = this.loadNewItem(this.header.appendChild(div), {
			$bind: "kind",
			$isEditMode: true,
			$format: "$combo",
			$inplace: true
		});
		var btn = syra_menus.addTextButton(syra_local.jobs_clear_list, "s-jobs-clear", "onClear");
		btn.syraItem = this.id;
		this.header.appendChild(btn);

		this.scrollview = document.createElement("div");
		this.scrollview.className = "s-jobs-body";
		this.domItem.appendChild(this.scrollview);
	},
	_showMessage: function($bind, dataset, cbck) {
		this.msgBoxId = syra_diagnose.showBox({
			$message: (syra_local['job_' + $bind.slice(1) + '_msg'].replace("{title}", dataset.title)).replace("{jobKind}", dataset.kind),
			$title: syra_local['job_' + $bind.slice(1) + '_title'],
			$type: "warning",
			$buttons: "yesno",
			$isAutoClose: 25000,
			$default: "no",
			callback: function(response, closedBy) {
				if (!(closedBy == "no" || closedBy == "auto")) {
					cbck();
				}
				return true;
			}
		}).id;
	},
	onClearJob: function(event, target) {
		var job = _jobs.map[target.syraJob];
		if (job.$progressValue == "on") {
			this._showMessage("$clearJob", job.dataset, function() {
				job.remove();
			});
		} else {
			job.remove();
		}
	},
	onClear: function(event, target, jobs) {
		var self = this;
		if (!jobs) {
			// get selected job type
			var selectedValue = self.kindField.currentValue;
			var $enum = self.kindField.$enum;
			var selectedTitle;
			for (var ii = 0, jj = $enum.length; ii < jj; ii++) {
				if ($enum[ii].$value == selectedValue) {
					selectedTitle = $enum[ii].$title;
					break;
				}
			}
			jobs = [];
			var onProcessing;
			for (var ii = 0, jj = _jobs.list.length; ii < jj; ii++) {
				var add = true;
				var job = _jobs.list[ii];
				if (selectedTitle && (job.dataset.kind != selectedTitle) && (selectedTitle != syra_local.job_kind_default)) {
					add = false;
				}
				if (add) {
					if (job.$progressValue == "on") {
						onProcessing = true;
					}
					jobs.push(job);
				}
			}
			if (onProcessing) {
				self._showMessage("$clearList", self.dataset, function() {
					self.onClear(event, target, jobs);
				});
			}
		}
		if (jobs) {
			while (jobs.length) {
				jobs.pop().remove(jobs.length == 0); //check kind on last clean
			}
		}
	},
	onJobActionClick: function(event, target) {
		var self = this;
		var job = _jobs.map[target.syraJob];
		var $menu = job.dataset.$links[target.syraBind];
		var $url = syra_site.urlMaker.formatMenuUrl(self, $menu, job.dataset);
		switch (target.syraBind) {
			case "$download":
				if ($menu.$verifyUrl) {
					var $verifyMenu = helpers.object.clone($menu, true);
					$verifyMenu.$sourceUrl = $verifyMenu.$url = $verifyMenu.$verifyUrl;
					delete $verifyMenu.$verifyUrl;
					var $verifyUrl = syra_site.urlMaker.formatMenuUrl(self, $verifyMenu, job.dataset);
					syra_controller.callServer(null, {
						noDisplayErr: true,
						method: "GET",
						$location: {
							$url: $verifyUrl
						},
						$acceptType: job.dataset.mime
					}, function() {
						window.open($url, "_blank");
					}, function(error, httpquery) {
						job.onRequestError(error, true);
					});
					return;
				}
				window.open($url, "_blank");
				break;
			case "$cancel":
				self._showMessage(target.syraBind, job.dataset, function() {
					job.onClientRelease();
				});
				break;
			case "$abort":
			case "$suspend":
			case "$resume":
				self._showMessage(target.syraBind, job.dataset, function() {
					job.onClientRelease($url, $menu.$method);
				});
				break;
			default:
				var win = window.open($url, '_self');
				win.focus();
				break;
		}
	},
	notifyDataChange: function(field, value) {
		for (var ii = 0, jj = field.$enum.length; ii < jj; ii++) {
			if (field.$enum[ii].$value == value) {
				var title = field.$enum[ii].$title;
				for (var mm = 0, pp = _jobs.list.length; mm < pp; mm++) {
					var job = _jobs.list[mm];
					job.domItem.style.display = ((job.dataset.kind == title) || (title == syra_local.job_kind_default)) ? "" : "none";
				}
				break;
			}
		}
	},
	checkKindList: function() {
		var kinds = {};
		var $enum = [];
		var curValue = this.kindField.getDataValue();
		for (var ii = 0, jj = _jobs.list.length; ii < jj; ii++) {
			var dataset = _jobs.list[ii].dataset;
			if (!kinds[dataset.kind]) {
				kinds[dataset.kind] = dataset.kind;
				$enum.push({
					$value: dataset.kind,
					$title: dataset.kind
				});
			}
		}
		if ($enum.length != 1) {
			$enum.push({
				$value: syra_local.job_kind_default,
				$title: syra_local.job_kind_default
			});
		}
		if (!kinds[curValue]) {
			curValue = null;
		}
		if (!curValue) {
			curValue = $enum.length == 1 ? $enum[0].$value : syra_local.job_kind_default;
		}
		this.kindField.applyMetaData({
			$value: {
				$enum: $enum
			}
		});
		this.kindField.setDataValue(curValue);
	},
	close: function() {

	},
	toggle: function(show) {
		var self = this;
		show = show || !_popup;
		if (self.domItem) {
			if (show) {
				self.domItem.style.display = "";
				syra_site.dom.setZIndex(self.domItem);
				if (!_popup) {
					self.layoutSlot.style.display = "";
					_popup = syra_site.dialogManager.openPopup(self, {
						content: self,
						$isAutoClose: false,
						slot: self.layoutSlot,
						picker: _opener,
						position: {
							my: "right top",
							at: "right bottom",
							of: $(_opener)
						},
						onClose: function() {
							setTimeout(function() {
								syra_site.dom.display(self.layoutSlot, false);
								_popup = null;
							}, 200);
						}
					});
					self.dialogWrapper = _popup;
					_popup.resizeDialog();
				}
			} else {
				_popup && _popup.close();
			}
		}
	},
	resizeArticle: function() {
		if (_popup) {
			var height = syra_site.body.getBoundingClientRect().height;
			height -= (syra_site.header && syra_site.headerTop.getBoundingClientRect().height) || 0;
			this.layoutSlot.style.height = "";
			this.layoutSlot.style.maxHeight = height + "px";
			var maxH = (height - this.header.getBoundingClientRect().height);
			this.scrollview.style.maxHeight = maxH + "px";
			var maxH;
			_popup.setPosition();
			for (var ii = 0, jj = _jobs.list.length; ii < jj; ++ii) {
				var job = _jobs.list[ii];
				if (job.diagnoses && job.diagnoses.slot) {
					job.diagnoses.slot.style.maxHeight = maxH + "px";
				}
			}
		}
	}
});


exports.addOpener = function(slot) {
	_opener = syra_menus.addIconButton("", "s-jobs-opener", "onJobsOpenerClick");
	_opener.syraItem = syra_site.id;
	syra_site.onJobsOpenerClick = function() {
		_page && _page.toggle();
	};
	_opener.style.display = "none";
	slot.appendChild(_opener);
};

exports.dispose = function() {
	_page && _page.dispose();
	_opener = _page = _popup = null;
};

function _runJobsViewer() {
	syra_controller.callServer(null, {
		method: "GET",
		$location: {
			$url: "/sdata/$trackers"
		}
	}, function(data, response, $url) {
		if (data.$resources && data.$resources.length > 0) {
			for (var ii = 0, jj = data.$resources.length; ii < jj; ii++) {
				var resource = data.$resources[ii];
				var params = {
					$location: resource.location || resource.$location,
					$state: resource.phase,
					$title: resource.title || "(Title not retrieved)",
					uuid: resource.uuid,
					kind: "operation"
				};
				syra_jobs.addJobOperation({
					$location: params.$location,
					$state: params.$state,
					title: params.$title,
					uuid: params.uuid
				});
			}
		}
	}, function(error, httpquery) {
		httpquery.showError(error);
	});
}


exports.addJobOperation = function(params) {
	params.kind = "operation";
	exports.addJob(params);
};
exports.addJob = function(dataset, fusionContext) {
	var job = new Job();
	job.fusionContext = fusionContext;
	job.dataset = dataset;
	job.dataset.$uuid = dataset.$uuid || dataset.uuid;
	_jobs.add(job);
	job.diagnoses = {
		items: []
	};
	job.draw();

	job.applyChange(dataset);
	if (job.fusionContext) {
		job.fusionContext.onLoaded(job);
	} else {
		job.callServer({
			noDisplayErr: true,
			method: "GET",
			$location: {
				$url: job.dataset.$location
			}
		});
	}
	_page.checkKindList();
};
