// syracuse-ui/lib/field/array/designer/cardDesigner.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var PageDesigner = require("syracuse-ui/lib/authoring/pageDesigner").PageDesigner;

function _toogleOverLaySlots(designer, show) {
	var ids = ["left", "top", "right", "bottom"];
	if (show) {
		this.overlays = {};
		for (var ii = 0, jj = ids.length; ii < jj; ii++) {
			var overlay = designer.overlays[ids[ii]] = document.createElement("div");
			overlay.syraIsOverlay = true;
			overlay.className = "s-aw-carddesigner-overlay";
			designer.designedArticle.page.scrollview.appendChild(overlay);
		}
	} else {
		if (designer.overlays) {
			for (var ii = 0, jj = ids.length; ii < jj; ii++) {
				syra_site.dom.removeChild(designer.overlays[ids[ii]]);
			}
			designer.overlays = null;
		}
	}
}


function CardDesigner() {}

exports.CardDesigner = helpers.defineClass(CardDesigner, PageDesigner, {
	drawBox: function() {
		PageDesigner.prototype.drawBox.call(this);
		syra_site.dom.removeChild(this.viewsSlot);
		this.viewsSlot = null;
		this.cardTitle = document.createElement("div");
		this.cardTitle.className = "s-aw-carddesigner-title";
		var title = this.designedArticle.list.getTitle() || "";
		if (title.length) {
			title += " - ";
		}
		title += this.designedArticle.list.$item.$format == "grid" ? syra_local.aw_card_detail_title : syra_local.aw_card_card_title;
		this.cardTitle.textContent = title;
		this.leftTop.className += " s-aw-carddesigner-left-top";
		this.layoutSlot.insertBefore(this.cardTitle, this.leftTop.nextSibling);
		_toogleOverLaySlots(this, true);
	},

	endArticleUpdate: function(selectedItem, isStructureUpdated) {
		PageDesigner.prototype.endArticleUpdate.call(this, selectedItem, isStructureUpdated);
		this.resizeArticle();
	},
	resizeArticle: function(resize) {
		PageDesigner.prototype.resizeArticle.call(this, resize);
		if (this.overlays && this.designedArticle.domItem) {
			var rectHeader = this.designedArticle.page.header.getBoundingClientRect();
			var rect = this.designedArticle.page.body.getBoundingClientRect();
			var rectCard = this.designedArticle.domItem.getBoundingClientRect();

			var top = rectHeader.height + (rectCard.top - rect.top);
			this.overlays.left.style.top = top + "px";
			this.overlays.left.style.left = 0;
			this.overlays.left.style.width = (rectCard.left - rect.left) + "px";
			this.overlays.left.style.height = rectCard.height + "px";

			this.overlays.right.style.top = top + "px";
			this.overlays.right.style.right = 0;
			this.overlays.right.style.width = (rect.right - rectCard.right) + "px";
			this.overlays.right.style.height = rectCard.height + "px";

			this.overlays.top.style.top = 0;
			this.overlays.top.style.left = 0;
			this.overlays.top.style.width = rect.width + "px";
			this.overlays.top.style.height = top + "px";

			this.overlays.bottom.style.top = (top + rectCard.height) + "px";
			this.overlays.bottom.style.left = 0;
			this.overlays.bottom.style.width = rect.width + "px";
			this.overlays.bottom.style.height = (rect.bottom - rectCard.bottom) + "px";
		}
	},
	openDesigner: function(designedArticle) {
		designedArticle.$designLevel = "article";
		PageDesigner.prototype.openDesigner.call(this, designedArticle);
	},
	closeDesigner: function() {
		if (this.history.currentStep > 1) {
			this.designedArticle.list.designer.history.updateSteps(true);
		}
		this.designedArticle.list.designer.designCardItem(false, this.designedArticle);
	},
	dispose: function() {
		_toogleOverLaySlots(this, false);
		PageDesigner.prototype.dispose.call(this);
	}
});
