// syracuse-ui/lib/tablet/hacksDadim.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
// !! Put the required in the functions by class instead of header because required could be buggy
var util = require('syracuse-ui/lib/fusion/core/client/sap/util');


function WorkBook() {

	var WorkBook = require('syracuse-ui/lib/fusion/workBook').WorkBook;
	var _onFieldEvent = WorkBook.prototype.onFieldEvent;
	WorkBook.prototype.onFieldEvent = function(options, redirect) {
		if (options && options.event && options.event.type === "focusin" && options.event.isSimulated && options.event.relatedTarget) {
			// simulate tabulation for the "next " button on tablet
			var e = jQuery.extend(new jQuery.Event(), {
				type: "keydown",
				//type: "mousedown",
				//target: options.event.target,
				target: options.event.relatedTarget,
				keyCode: keys["VK_TAB"],
				isSimulated: true,
				originalEvent: {}
			});

			options.event = e;


		}

		return _onFieldEvent.call(this, options, redirect);
	};
}


exports.init = function() {
	WorkBook();
};
