// syracuse-ui/lib/tablet/hacks.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var NumericField = require('syracuse-ui/lib/field/numericField').NumericField;
var MenuItem = require('syracuse-ui/lib/menus/menuItem').MenuItem;

var CommonEventHandlingTools = require('syracuse-ui/lib/fusion/core/client/eventhandling/common').CommonEventHandlingTools;

function numericField() {
	// Display numeric panel -> Type should be number
	var render = NumericField.prototype.render;
	NumericField.prototype.render = function() {
		var res = render.call(this);
		this._input && this._input.setAttribute("type", "number");
		return res;
	};
}

function menuItem() {
	// Hide excel and word links in right panel
	var loadBox = MenuItem.prototype.loadBox;
	MenuItem.prototype.loadBox = function() {
		var res = loadBox.call(this);
		if (this.$item.$bind && typeof this.$item.$bind === "string" &&
			(this.$item.$bind.indexOf("$word") === 0 ||
				this.$item.$bind.indexOf("$excel") === 0 ||
				this.$item.$bind.indexOf("$pptslide") === 0)) {

			this._hide(true);
		}
		console.log(this.$item && this.$item.$bind);
		return res;
	};
}

function commonEventHandlingTools() {
	CommonEventHandlingTools._checkCtxMenuEntry = function(men) {
		if (men && men.id && men.id === "$tabularExport")
			return false;
		return true;
	};
}

exports.init = function() {
	numericField();
	menuItem();
	commonEventHandlingTools();
};
