// syracuse-ui/lib/tablet/hacks.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function menuItem() {
	var MenuItem = require('syracuse-ui/lib/menus/menuItem').MenuItem;
	var MenuGroup = require('syracuse-ui/lib/menus/menuGroup').MenuGroup;
	// Hide excel and word links in right panel
	var loadBox = MenuItem.prototype.loadBox;
	MenuItem.prototype.loadBox = function() {
		var res = loadBox.call(this);
		var hidden = false;
		// syracuse
		// console.log(this.$item.$bind + "/" + (this.$type && typeof this.$type));
		if (this.$item.$bind && typeof this.$item.$bind === "string" &&
			(this.$item.$bind.indexOf("$word") === 0 ||
				this.$item.$bind.indexOf("$excel") === 0 ||
				this.$item.$bind.indexOf("$pptslide") === 0)) {
			hidden = true;
		}
		// cvg by type
		else
		if (this.$type && typeof this.$type === "string" &&
			(this.$type.indexOf("application/syracuse-excel-worksheet") >= 0 ||
				this.$type.indexOf("application/syracuse-word-report") >= 0 ||
				this.$type.indexOf("application/syracuse-word-mailmerge") >= 0 ||
				this.$type.indexOf("application/syracuse-ppt-slide") >= 0)) {
			hidden = true;
		}
		// cvg by special name
		else
		if (this.$item.$bind && typeof this.$item.$bind === "string" &&
			(this.$item.$bind.indexOf("cvgOffice") === 0)) {
			hidden = true;
		}

		if (hidden === true) {
			this.$isHidden = true;
			this.showItem(!this.$isHidden);
		}
		return res;
	};

	/*
     Hide utilities menu
     */
	function isMenuGroupHidden($item) {
		var ids = ["318", // SUBMENU_DIAGNOS_CMN
			"40003", // Global export
			"18000" //Export
		];
		return ($item && $item.$clientId && ids.indexOf($item.$clientId) >= 0) ? true : false;
	}
	var menuGroupLoadBox = MenuGroup.prototype.loadBox;
	var menuGroupShowItem = MenuGroup.prototype.showItem;
	MenuGroup.prototype.loadBox = function() {
		var res = menuGroupLoadBox.call(this);
		if (isMenuGroupHidden(this.$item)) {
			if (this.domItem) {
				this.$isHidden = true;
				this.showItem(false);
			}
		}
		return res;
	};
	MenuGroup.prototype.showItem = function(show) {
		if (isMenuGroupHidden(this.$item)) {
			show = false;
		}
		var res = menuGroupShowItem.call(this, show);
		return res;
	};
}

function commonEventHandlingTools() {
	var CommonEventHandlingTools = require('syracuse-ui/lib/fusion/core/client/eventhandling/common').CommonEventHandlingTools;
	CommonEventHandlingTools._checkCtxMenuEntry = function(men) {
		if (men && men.id && men.id === "$tabularExport")
			return false;
		return true;
	};
}

function removeFlash() {
	var flashWidget = require('syracuse-ui/lib/field/chart/flashWidget');
	flashWidget.FlashWidget.prototype.drawBox = function() {
		var $cnt = $("<div class=\"s-h1-body s-cst-sty-not-sup\">" + syra_local.table_not_sup + "</div>");
		$(this.layoutSlot).append($cnt);
	};
	flashWidget.FlashWidget.prototype.resizeChart = function() {};
	flashWidget.FlashWidget.prototype.setDataBind = function(dataRecordSet, parentDataRecord, metaData) {};
	flashWidget.FlashWidget.prototype.applyDesignMetaData = function(metadata, designing) {};
	flashWidget.FlashWidget.prototype.dispose = function() {};

	flashWidget.CubeChart.prototype.drawBox = function() {
		var $cnt = $("<div class=\"s-h1-body s-cst-sty-not-sup\">" + syra_local.table_not_sup + "</div>");
		$(this.layoutSlot).append($cnt);
	};
	flashWidget.CubeChart.prototype.resizeChart = function() {};
	flashWidget.CubeChart.prototype.setDataBind = function(dataRecordSet, parentDataRecord, metaData) {};
	flashWidget.CubeChart.prototype.applyDesignMetaData = function(metadata, designing) {};
	flashWidget.CubeChart.prototype.dispose = function() {};
}
exports.initAfterLoad = function(site) {
	menuItem();
	commonEventHandlingTools();
	removeFlash();
};
