// syracuse-ui/lib/tablet/tabletSite.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var MobileEventListener = require('./mobileEventListener').MobileEventListener;
var Site = require("syracuse-ui/lib/site/site").Site;
//var WidgetResizer = require("syracuse-ui/lib/site/resize/widgetResizer").WidgetResizer;
//var ResizeListener = require("syracuse-ui/lib/site/resize/resizeListener").ResizeListener;

var TabletWidgetResizer = require("syracuse-ui/lib/tablet/resize/widgetResizer").TabletWidgetResizer;
var ResizeListener = require("syracuse-ui/lib/tablet/resize/resizeListener").ResizeListener;



function TabletSite() {}

exports.TabletSite = helpers.defineClass(TabletSite, Site, {
	unload: function() {

		//document.controller.disposeObject(this._eventListener);

	},
	//, this.isMobile : true ;
	setWidgetResizable: function(options) {
		var widgetResizer = new TabletWidgetResizer();
		//widgetResizer.setResizable(options);
		return widgetResizer;
	},
	ensurePreferences: function() {
		return (this._preferences = this._preferences || {
			convergenceBar: {
				isDocked: true,
				isCollapsed: true
			}
		});
	},
	/*,
	attachResizeListeners: function(attach, page) {
		if (page) {
			if (!page._unregisterResizer) {
				var parent = page.dialogWrapper || document.site;
				if (parent && parent.resizeListener) {
					if (attach) {
						parent.resizeListener.add(page);
					} else {
						parent.resizeListener.remove(page);
					}
				}
			}
		}
		if (attach) {
			(page.resizeListener = new ResizeListener()).load();
		} else {
			if (page.resizeListener) {
				page.resizeListener.dispose();
				page.resizeListener = null;
			}
		}

	}*/
	/*resize: function(onBoxToggle) {
		var self = this;
		self.resizeTimeOut = setTimeout(function() {
			if (self.$$body && self.body) {
				if (self.authorPage && self.authorPage.onBeforeSiteResize) {
					self.authorPage.onBeforeSiteResize(self);
				} else {
					var newHeight = 0;
					if (self.header) {
						self._$$header = self._$$header || $(self.header);
						newHeight = self._$$header.outerHeight(true);
					}
					self.body.style.height = (self.$$layoutSlot.height() - newHeight) + "px";
				}
			}
			//self.resizeListener.notifyObservers(onBoxToggle);
			clearTimeout(self.resizeTimeOut);
			self.ensureUILocker();
		}, 10);
	}*/

});


exports.load = function($item, $prototype) {
	//(this._eventListener = new EventListener()).load();

	return (new TabletSite()).loadBox({
		widgetsLibrary: require('syracuse-ui/lib/site/widgetsLibrary'),
		fusionGatewayClass: require('syracuse-ui/lib/fusion/fusionGateway').FusionGateway,
		userProfileClass: require('syracuse-ui/lib/site/userProfile').UserProfile,
		mobileEventListener: require('./mobileEventListener').MobileEventListener,
		$item: $item,
		$prototype: $prototype
	});
};
