// @sage/syracuse-react/src/common/sortable/sortableLine.tsx
"use strict";
import * as React from 'react';
import * as ReactDom from 'react-dom';
import { Icon } from '../icon/icon';

export class SortableLine extends React.Component<{
    id?:string,
    selected?: boolean,
    className: string,
    value: any,
    onSelect?: (value: any) => void,
    onStartSort: (value: any, line: HTMLElement) => void
}, {}>{
    handleClick = () => {
        this.props.onSelect && this.props.onSelect(this.props.value);
    }
    handleMouseDown = () => {
        this.props.onStartSort(this.props.value, ReactDom.findDOMNode(this) as HTMLElement);
    }
    render() {
        return <li id={this.props.id} className={this.props.className + (this.props.selected ? " s_selected" : "")} onClick={this.handleClick}>
            <div className={this.props.className + "_sortable_handler s_sortable_handler"} onMouseDown={this.handleMouseDown}>
                <Icon name="move_1" />                
            </div>
            {this.props.children}
        </li>;
    }
}
