// syracuse-ui/lib/site/legal/legalPageOpener.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var LegalPage = require('./legalPage').LegalPage;

var $legalDataUrl = "/sdata/syracuse/collaboration/syracuse/licenseDatas?representation=licenseData.$query";
var $legalPrototypeUrl = "/sdata/syracuse/collaboration/syracuse/$prototypes('licenseData.$query')";

exports.openLegalPage = function(menuItem) {

	document.controller.sendRequest(null, {
		method: "GET",
		$location: {
			$url: $legalPrototypeUrl
		}
	}, function(data, response, $url) {
		var legalPage = new LegalPage();
		legalPage.loadBox(data);
		if (legalPage.menuBar) {
			legalPage.menuBar.toggleBar(false);
		}
		document.controller.sendRequest(null, {
			method: "GET",
			$location: {
				$url: $legalDataUrl
			}
		}, function(data, response, $url) {
			legalPage.applyChange(data);
			var options = {
				page: legalPage
			};
			document.site.dialogManager.openModal(menuItem.page, options);
		}, function(error) {
			var $diagnoses;
			if (error.data.indexOf("$diagnoses") != -1) {
				$diagnoses = JSON.parse(error.data).$diagnoses;
			} else {
				$diagnoses = [{
					$severity: "error",
					$message: error.data
				}];
			}
			document.site.showDiagnoses({
				$diagnoses: $diagnoses
			}, page);
		});

	}, function(error) {
		var $diagnoses;
		if (error.data.indexOf("$diagnoses") != -1) {
			$diagnoses = JSON.parse(error.data).$diagnoses;
		} else {
			$diagnoses = [{
				$severity: "error",
				$message: error.data
			}];
		}
		document.site.showDiagnoses({
			$diagnoses: $diagnoses
		}, page);
	});

};
