// syracuse-ui/lib/tests/pages/misc/office.js
"use strict";
var _$groups = [{
	$title: "Office & Documents",
	$menus: [{
		$title: "Prints",
		$url: "/sdata/syracuse/collaboration/syracuse/printDocuments?representation=printDocument.$query"
	}, {
		$title: "Word template documents",
		$url: "/sdata/syracuse/collaboration/syracuse/msoWordTemplateDocuments?representation=msoWordTemplateDocument.$query",
	}, {
		$title: "Excel template documents",
		$url: "/sdata/syracuse/collaboration/syracuse/msoExcelTemplateDocuments?representation=msoExcelTemplateDocument.$query",
	}]
}, {
	$title: "Excel test pages",
	$menus: [{
		$title: "Excel settings dashboard",
		$url: "/msoffice/lib/excel/html/main.html?url=?representation=excelconfig.$query",
		$type: "application/syracuse-excel-worksheet"
	}, {
		"$title": "Excel action panel",
		"$url": "/msoffice/lib/excel/html/main.html?url=%3Frepresentation%3Dexcelconfig.%24details",
		$type: "application/syracuse-excel-worksheet"
	}, {
		$title: "Excel test Template",
		$url: "/msoffice/lib/excel/html/main.html?url=%3Frepresentation%3Dexceltemplate.%24query",
		$type: "application/syracuse-excel-worksheet"
	}, {
		$title: "Convergence export test",
		$url: "/sdata/syracuse/collaboration/syracuse/users?representation=user.%24query&format=application/syracuse-excel-worksheet&testStubFile=cvgExcelStub1.json",
		$type: "application/syracuse-excel-worksheet"
	}, {
		$title: "Convergence export test (empty dates)",
		$url: "/sdata/syracuse/collaboration/syracuse/users?representation=user.%24query&format=application/syracuse-excel-worksheet&testStubFile=cvgExcelStub2.json",
		$type: "application/syracuse-excel-worksheet"
	}, {
		$title: "Convergence export test (one row)",
		$url: "/sdata/syracuse/collaboration/syracuse/users?representation=user.%24query&format=application/syracuse-excel-worksheet&testStubFile=cvgExcelStub3.json",
		$type: "application/syracuse-excel-worksheet"
	}, {
		$title: "GESPOH Lines (Cell formats)",
		$url: "/sdata/syracuse/collaboration/syracuse/users?representation=user.%24query&format=application/syracuse-excel-worksheet&testStubFile=cvgExcelStub4.json",
		$type: "application/syracuse-excel-worksheet"
	}, {
		$title: "Test 5",
		$url: "/sdata/syracuse/collaboration/syracuse/users?representation=user.%24query&format=application/syracuse-excel-worksheet&testStubFile=cvgExcelStub5.json",
		$type: "application/syracuse-excel-worksheet"
	}]
}, {
	$title: "Word test pages",
	$menus: [{
		$title: "Word test Template",
		$url: "/msoffice/lib/word/html/main.html?url=%3Frepresentation%3Dwordtemplate.%24query",
		$type: "application/syracuse-word-report"
	}, {
		$title: "Report template (user.$details)",
		$type: "application/syracuse-word-report",
		$url: "/sdata/syracuse/collaboration/syracuse/users(login eq 'admin')?representation=user.$details&reportMode=new_tpl&format=application/syracuse-word-report",
		$target: "blank"
	}, {
		$title: "Report template (user.$query)",
		$type: "application/syracuse-word-report",
		$url: "/sdata/syracuse/collaboration/syracuse/users?representation=user.$query&reportMode=new_tpl&format=application/syracuse-word-report",
		$target: "blank"

	}, {
		$title: "Report document DE-de soTestEntity.$query)",
		$type: "application/syracuse-word-report",
		$url: "/sdata/syracuse/collaboration/syracuse/msoTestEntities?representation=msoTestEntity.$query&reportMode=choose_tpl&format=application/syracuse-word-report&test_template=test_page_1",
		$target: "blank"
	}, {
		$title: "Report document EN-us (msoTestEntity.$query)",
		$type: "application/syracuse-word-report",
		$url: "/sdata/syracuse/collaboration/syracuse/msoTestEntities?representation=msoTestEntity.$query&reportMode=choose_tpl&format=application/syracuse-word-report&test_template=test_page_2",
		$target: "blank"
	}, {
		$title: "Report data (msoTestEntity.$query)",
		$url: "http://localhost:8124/msoffice/lib/word/ui/main.html?url=%2Fsdata%2Fsyracuse%2Fcollaboration%2Fsyracuse%2FmsoTestEntities%3Frepresentation%3DmsoTestEntity.%24query&mode=word-report",

	}, {
		$title: "Report data (user.$query)",
		$url: "http://localhost:8124/msoffice/lib/word/ui/main.html?url=%2Fsdata%2Fsyracuse%2Fcollaboration%2Fsyracuse%2Fusers%3Frepresentation%3Duser.%24query&mode=word-report",

	}, {
		$title: "Report data (user.$details)",
		$url: "http://localhost:8124/msoffice/lib/word/ui/main.html?url=http%3A%2F%2Flocalhost%3A8124%2Fsdata%2Fsyracuse%2Fcollaboration%2Fsyracuse%2Fusers(login eq 'admin')%3Frepresentation%3Duser.%24details&mode=word-report",
	}, {
		$title: "Report data (msoTestEntity.$query)",
		$url: "/sdata/syracuse/collaboration/syracuse/msoTestEntities?representation=msoTestEntity.$query",
	}]
}, {
	$title: "PPT test pages",
	$menus: [{
		$title: "PPT Spiderweb (Preview)",
		$url: "/sdata/x3stb/erp/stubs/pptHighCharts1?representation=pptHighCharts1.$query",
		$target: "blank"
	}, {
		$title: "PPT Spiderweb",
		$type: "application/syracuse-ppt-slide",
		$url: "/sdata/x3stb/erp/stubs/pptHighCharts1?representation=pptHighCharts1.$query&pptMode=newSlide&format=application/syracuse-ppt-slide&$bind=A",
		$target: "blank"
	}, {
		$title: "PPT Spiderweb (Data)",
		$url: "http://localhost:8124/msoffice/lib/ppt/ui/main.html?url=%2Fsdata%2Fx3stb%2Ferp%2Fstubs%2FpptHighCharts1%3Frepresentation%3DpptHighCharts1.%24query%26%24bind=A",
		$target: "blank"
	}, {
		$title: "PPT Columns (Preview)",
		$url: "/sdata/x3stb/erp/stubs/pptHighCharts2?representation=pptHighCharts2a.$query",
		$target: "blank"
	}, {
		$title: "PPT Columns (Data)",
		$url: "http://localhost:8124/msoffice/lib/ppt/ui/main.html?url=%2Fsdata%2Fx3stb%2Ferp%2Fstubs%2FpptHighCharts2%3Frepresentation%3DpptHighCharts2.%24query%26%24bind=A",
		$target: "blank"
	}, {
		$title: "PPT Columns",
		$type: "application/syracuse-ppt-slide",
		$url: "/sdata/x3stb/erp/stubs/pptHighCharts2?representation=pptHighCharts2.$query&pptMode=newSlide&format=application/syracuse-ppt-slide&$bind=A",
		$target: "blank",
	}, {
		$title: "PPT Columns, Stacked (Preview)",
		$url: "/sdata/x3stb/erp/stubs/pptHighCharts2?representation=pptHighCharts2b.$query",
		$target: "blank",
	}, {
		$title: "PPT Columns, Stacked (Data)",
		$url: "http://localhost:8124/msoffice/lib/ppt/ui/main.html?url=%2Fsdata%2Fx3stb%2Ferp%2Fstubs%2FpptHighCharts2%3Frepresentation%3DpptHighCharts2.%24query%26%24bind=B",
		$target: "blank",
	}, {
		$title: "PPT Columns, Stacked",
		$type: "application/syracuse-ppt-slide",
		$url: "/sdata/x3stb/erp/stubs/pptHighCharts2?representation=pptHighCharts2.$query&pptMode=newSlide&format=application/syracuse-ppt-slide&$bind=B",
		$target: "blank",
	}, {
		$title: "PPT Columns, Stacked, Normalized (Preview)",
		$url: "/sdata/x3stb/erp/stubs/pptHighCharts2?representation=pptHighCharts2c.$query",
		$target: "blank",
	}, {
		$title: "PPT Columns, Stacked, Normalized (Data)",
		$url: "http://localhost:8124/msoffice/lib/ppt/ui/main.html?url=%2Fsdata%2Fx3stb%2Ferp%2Fstubs%2FpptHighCharts2%3Frepresentation%3DpptHighCharts2.%24query%26%24bind=C",
		$target: "blank",
	}, {
		$title: "PPT Columns, Stacked, Normalized",
		$type: "application/syracuse-ppt-slide",
		$url: "/sdata/x3stb/erp/stubs/pptHighCharts2?representation=pptHighCharts2.$query&pptMode=newSlide&format=application/syracuse-ppt-slide&$bind=C",
		$target: "blank",
	}, {
		$title: "PPT Columns, Stacked, Normalized, Grouped (Preview)",
		$url: "/sdata/x3stb/erp/stubs/pptHighCharts2?representation=pptHighCharts2d.$query",
		$target: "blank",
	}, {
		$title: "PPT Columns, Stacked, Normalized, Grouped (Data)",
		$url: "http://localhost:8124/msoffice/lib/ppt/ui/main.html?url=%2Fsdata%2Fx3stb%2Ferp%2Fstubs%2FpptHighCharts2%3Frepresentation%3DpptHighCharts2.%24query%26%24bind=D",
		$target: "blank",
	}, {
		$title: "PPT Columns, Stacked, Normalized, Grouped",
		$type: "application/syracuse-ppt-slide",
		$url: "/sdata/x3stb/erp/stubs/pptHighCharts2?representation=pptHighCharts2.$query&pptMode=newSlide&format=application/syracuse-ppt-slide&$bind=D",
		$target: "blank",
	}, {
		$title: "PPT Area Spline, Stacked, Normalized (Preview)",
		$url: "/sdata/x3stb/erp/stubs/pptHighCharts2?representation=pptHighCharts2e.$query",
		$target: "blank",
	}, {
		$title: "PPT Area Spline, Stacked, Normalized (Data)",
		$url: "http://localhost:8124/msoffice/lib/ppt/ui/main.html?url=%2Fsdata%2Fx3stb%2Ferp%2Fstubs%2FpptHighCharts2%3Frepresentation%3DpptHighCharts2.%24query%26%24bind=E",
		$target: "blank",
	}, {
		$title: "PPT Area Spline, Stacked, Normalized ",
		$type: "application/syracuse-ppt-slide",
		$url: "/sdata/x3stb/erp/stubs/pptHighCharts2?representation=pptHighCharts2.$query&pptMode=newSlide&format=application/syracuse-ppt-slide&$bind=E",
		$target: "blank",
	}, {
		$title: "PPT Chart Test 3 (Preview)",
		$url: "/sdata/x3stb/erp/stubs/pptHighCharts3?representation=pptHighCharts3.$query",
		$target: "blank",
	}, {
		$title: "PPT Chart Test 3 (Data)",
		$url: "http://localhost:8124/msoffice/lib/ppt/ui/main.html?url=%2Fsdata%2Fx3stb%2Ferp%2Fstubs%2FpptHighCharts3%3Frepresentation%3DpptHighCharts3.%24query%26pptMode=newSlide%26%24bind=%24resources",
		$target: "blank",
	}, {
		$title: "PPT Chart Test 3",
		$type: "application/syracuse-ppt-slide",
		$url: "/sdata/x3stb/erp/stubs/pptHighCharts3?representation=pptHighCharts3.$query&pptMode=newSlide&format=application/syracuse-ppt-slide&$bind=$resources",
		$target: "blank",
	}]
}];



exports.load = function(page) {
	page.setTitle("Office");
	var $links = {};
	for (var ii = 0, jj = _$groups.length; ii < jj; ii++) {
		var $group = _$groups[ii];
		var box = page.addBox($group.$title, "h2");
		for (var mm = 0, kk = $group.$menus.length; mm < kk; mm++) {
			var $mn = $group.$menus[mm];
			var id = page.id + "-" + ii + "-" + mm;
			$links[id] = $mn;
			page.loadNewItem(box.body, {
				$category: "link",
				$skin: "s-uitest-mn",
				$bind: id
			});
		}
		((ii < 3) ? page.leftCol : page.rightCol).appendChild(box.slot);
	}
	page.applyChange({
		$links: $links
	});
};
