// syracuse-ui/lib/tests/pages/misc/parameters.js
"use strict";
exports.load = function(page, finish) {
	page.applyChange({
		linkParameterField: JSON.stringify({
			$links: page.$prototype.$links.linkParameter
		}, null, 2),
		linkFormParameterField: JSON.stringify({
			$links: page.$prototype.$links.linkFormParameter
		}, null, 2),
		actionParameterField: JSON.stringify({
			$links: page.$prototype.$actions.actionParameter
		}, null, 2),
		actionFormParameterField: JSON.stringify({
			$links: page.$prototype.$actions.actionFormParameter
		}, null, 2),
		actionSelectFormParameterField: JSON.stringify({
			$links: page.$prototype.$actions.actionSelectFormParameter
		}, null, 2)
	});
	page.onMenuClick = function(options) {
		if (options.menu.$bind.indexOf("Parameter") > 0) {
			syra_url.formatMenuUrl(page, options.menu, null, function($url, isCanceled) {
				var result = {};
				result[options.menu.$bind + "FieldResult"] = JSON.stringify({
					actionFormParameterField: $url
				}, null, 2);
				page.applyChange(result);
			});
			return false;
		}
		return true;
	};
	page.updateWorkingCopy = function(sendBag) {
		this.applyChange(sendBag);
	};
	page.setTitle("Dynamic Paramerters");
	page.$prototype.$properties = {
		"fieldA": {
			"$title": "fieldA",
			"$type": "application/x-string"
		},
		"fieldB": {
			"$title": "fieldB",
			"$type": "application/x-string"
		},
		linkParameterField: {
			$title: "json definition",
			$type: "application/x-string"
		},
		linkParameterFieldResult: {
			$title: "result",
			$type: "application/x-string"
		},
		linkFormParameterField: {
			$title: "json definition",
			$type: "application/x-string"
		},
		linkFormParameterFieldResult: {
			$title: "result",
			$type: "application/x-string"
		},
		actionParameterField: {
			$title: "json definition",
			$type: "application/x-string"
		},
		actionParameterFieldResult: {
			$title: "result",
			$type: "application/x-string"
		},
		actionFormParameterField: {
			$title: "json definition",
			$type: "application/x-string"
		},
		actionFormParameterFieldResult: {
			$title: "result",
			$type: "application/x-string"
		},
		actionSelectFormParameterField: {
			$title: "json definition",
			$type: "application/x-string"
		}
	};
	page.$prototype.fieldA = "fieldA value";
	page.$prototype.fieldB = "fieldB value";

	page.$prototype.$links.$help = {
		$url: "https://github.com/Sage-ERP-X3/platform/wiki/Dynamic-parameters"
	};
	page.$prototype.$links.linkParameter = {
		"$title": "clik link with parameters",
		"$method": "POST",
		"$url": "/sdata/syracuse/collaboration/syracuse/users?representation=user.$query&par1={param1}&par2={param2}",
		"$parameters": {
			"param1": "{fieldA}",
			"param2": "value"
		}
	};
	page.$prototype.$links.linkFormParameter = {
		"$title": "clik link with form parameters",
		"$url": "/sdata/syracuse/collaboration/syracuse/users?representation=user.$query&par1={param1}&par2={param2}&par3={param3}&par4={param4}&par4={param5}",
		"$parameters": {
			"$properties": {
				"param3": {
					"$title": "my param3",
					"$type": "application/x-string"
				},
				"param4": {
					"$title": "my param4",
					"$type": "application/x-string"
				},
				"param5": {
					"$title": "my param5",
					"$type": "application/x-string",
					"$isEditMode": false
				}
			},
			"param3": "param3 default value",
			"param1": "{fieldA}",
			"param2": "{fieldB}"
		}
	};
	page.$prototype.$actions.actionParameter = {
		"$title": "click action with parameters",
		"$url": "/sdata/syracuse/collaboration/syracuse/users?representation=user.$query",
		"$parameters": {
			"param1": "{fieldA}",
			"param2": "value"
		}
	};
	page.$prototype.$actions.actionFormParameter = {
		"$title": "click action with form parameters",
		"$url": "/sdata/syracuse/collaboration/syracuse/users?representation=user.$query",
		"$parameters": {
			"$properties": {
				"param3": {
					"$title": "my param3",
					"$type": "application/x-string"
				},
				"param4": {
					"$title": "my param4",
					"$type": "application/x-string"
				},
				"param5": {
					"$title": "my param4",
					"$type": "application/x-string",
					"$isEditMode": false
				}
			},
			"param3": "param3 default value",
			"param1": "{fieldA}",
			"param2": "{fieldB}"
		}
	};
	page.$prototype.$actions.actionSelectFormParameter = {
		"$title": "click action with select form parameters",
		"$url": "/sdata/syracuse/collaboration/syracuse/users?representation=user.$query",
		"$parameters": {
			"$actions": {
				"$select": {
					"$type": "application/x-array",
					"$url": "/sdata/syracuse/collaboration/syracuse/users?representation=user.$select"
				}
			}
		}
	};
	page.leftCol.style.width = "20%";
	page.rightCol.style.width = "80%";
	page.addItem(page.leftCol, {
		$bind: "fieldA",
		$isEditMode: true
	});
	page.addItem(page.leftCol, {
		$bind: "fieldB",
		$isEditMode: true
	});

	var box = page.addBox("Link Parameter", "h2");
	page.addItem(box.body, {
		$items: [{
			$category: "link",
			$bind: "linkParameter"
		}, {
			$bind: "linkParameterField",
			$isEditMode: true,
			$rows: 10,
			$isAutoSizeDisabled: true,
			$skin: "s-json-field"
		}, {
			$bind: "linkParameterFieldResult",
			$isEditMode: true,
			$rows: 10,
			$isAutoSizeDisabled: true,
			$skin: "s-json-field"
		}]
	});
	page.rightCol.appendChild(box.slot);


	var box = page.addBox("Link Form Parameter", "h2");
	page.addItem(box.body, {
		$items: [{
			$category: "link",
			$bind: "linkFormParameter"
		}, {
			$bind: "linkFormParameterField",
			$isEditMode: true,
			$rows: 28,
			$isAutoSizeDisabled: true,
			$skin: "s-json-field"
		}, {
			$bind: "linkFormParameterFieldResult",
			$isEditMode: true,
			$rows: 10,
			$isAutoSizeDisabled: true,
			$skin: "s-json-field"
		}]
	});
	page.rightCol.appendChild(box.slot);

	var box = page.addBox("Action Parameter", "h2");
	page.addItem(box.body, {
		$items: [{
			$category: "link",
			$bind: "actionParameter"
		}, {
			$bind: "actionParameterField",
			$isEditMode: true,
			$rows: 10,
			$isAutoSizeDisabled: true,
			$skin: "s-json-field"
		}, {
			$bind: "actionParameterFieldResult",
			$isEditMode: true,
			$rows: 10,
			$isAutoSizeDisabled: true,
			$skin: "s-json-field"
		}]
	});
	page.rightCol.appendChild(box.slot);


	var box = page.addBox("Action Form Parameter", "h2");
	page.addItem(box.body, {
		$items: [{
			$category: "link",
			$bind: "actionFormParameter"
		}, {
			$bind: "actionFormParameterField",
			$isEditMode: true,
			$rows: 28,
			$isAutoSizeDisabled: true,
			$skin: "s-json-field"
		}, {
			$bind: "actionFormParameterFieldResult",
			$isEditMode: true,
			$rows: 10,
			$isAutoSizeDisabled: true,
			$skin: "s-json-field"
		}]
	});
	page.rightCol.appendChild(box.slot);

	var box = page.addBox("Action Select Form", "h2");
	page.addItem(box.body, {
		$items: [{
			$category: "link",
			$bind: "actionSelectFormParameter"
		}, {
			$bind: "actionSelectFormParameterField",
			$isEditMode: true,
			$rows: 15,
			$isAutoSizeDisabled: true,
			$skin: "s-json-field"
		}, {
			$bind: "actionFormParameterFieldResult",
			$isEditMode: true,
			$rows: 10,
			$isAutoSizeDisabled: true,
			$skin: "s-json-field"
		}]
	});
	page.rightCol.appendChild(box.slot);

	finish(page);
};
