// syracuse-ui/lib/authoring/fieldAddins/htmlFielddAddin.js
"use strict";
exports.load = function(designer) {
	designer.$prototype.$properties.htmhField$title = {
		$type: "application/x-string",
		$title: syra_local.htmlVignetteTitle,
		$displayLength: 200
	};
	designer.$prototype.$properties.htmhField$url = {
		$type: "application/x-string",
		$isHidden: true,
		$title: syra_local.htmlVignetteUrl,
		$displayLength: 200
	};
	designer.$prototype.$properties.htmhField$height = {
		$type: "application/x-string",
		$title: syra_local.htmlVignetteHeight,
		$displayLength: 8
	};
};

exports.onShow = function(designer) {
	var item = designer.designedItem;
	if (item.$authoringType == "htmlVignetteField") {
		designer.applyChange({
			htmhField$title: item.getTitle() || item.$item.$title,
			htmhField$url: item.$contentUrl,
			htmhField$height: item.$item.$height
		});
	}
};

exports.onDesignerApplyChange = function(newData) {};

exports.notifyDataChange = function(designer, metaData) {
	if (designer.designedItem.$authoringType == "htmlVignetteField") {
		if (metaData.htmhField$title != undefined || metaData.htmhField$url != undefined || metaData.htmhField$height != undefined) {
			designer.designedItem.applyDesignMetaData({
				$title: metaData.htmhField$title,
				$url: metaData.htmhField$url,
				$height: metaData.htmhField$height
			}, true);
		}
	}
};

exports.addToSection = function(designer, section) {
	section.fields.push(designer.loadNewItem(section.body, {
		$authoringType: "htmlVignetteField",
		$bind: "htmhField$title",
		$isEditMode: true
	}, designer));
	section.fields.push(designer.loadNewItem(section.body, {
		$authoringType: "htmlVignetteField",
		$bind: "htmhField$url",
		$isEditMode: true
	}, designer));
	section.fields.push(designer.loadNewItem(section.body, {
		$authoringType: "htmlVignetteField",
		$bind: "htmhField$height",
		$isEditMode: true
	}, designer));
};
