// syracuse-ui/lib/field/choice/choiceFieldAuthoring.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var Article = require("syracuse-ui/lib/article/article").Article;
var locale = require('syracuse-core/lib/locale');


function ChoiceFieldAuthoring(){

}

exports.ChoiceFieldAuthoring = helpers.defineClass(ChoiceFieldAuthoring, Article, {
    notifyChange: function($bind, value,isNotifyOnly){
        var metaData = {};
        metaData[$bind] = value;
        this.designedField.applyDesignMetaData(metaData, true);
        if ($bind == "$format") {
            this.applyChange(metaData);
        }
        document.site.authorPage.notifyUpdate();
    },
    applyChange: function(newData){
        Article.prototype.applyChange.call(this, newData);
        if (newData) {
            var self = this;
            if (self.designedField.$item.$format == "$combo") {
                self.designedCombo = self.designedField;
                /*setTimeout(function(){
                 var builder = self.designedField ? self.designedField.builder : null;
                 if (builder) {
                 builder.onAuthoring = true;
                 builder.toggleList(true);
                 }
                 }, 20);*/
            }
            else {
                delete self.designedCombo;
            }
        }
    },
    loadBox: function(initData){
        var localize = locale.resources(module)();
        this.$prototype = {
            "$properties": {
                $format: {
                    $type: "application/x-choice",
                    $title: localize.cfa_format_title,
                    $value: {
                        $type: "application/x-string",
                        
                        $enum: [{
                            $value: "$combo",
                            $title: localize.cfa_format_enum_dropdownlist
                        }, {
                            $value: "$radios",
                            $title: localize.cfa_format_enum_radiobuttons
                        }]
                    }
                },
                $choiceLayout: {
                    $type: "application/x-choice",
                    $title: localize.cfa_choiceLayout_title,
                    $value: {
                        $type: "application/x-string",
                        
                        $enum: [{
                            $value: "row",
                            $title: localize.cfa_choiceLayout_enum_1_row
                        }, {
                            $value: "1",
                            $title: localize.cfa_choiceLayout_enum_1_column
                        }, {
                            $value: "2",
                            $title: localize.cfa_choiceLayout_enum_2_columns
                        }, {
                            $value: "3",
                            $title: localize.cfa_choiceLayout_enum_3_columns
                        }, {
                            $value: "4",
                            $title: localize.cfa_choiceLayout_enum_4_columns
                        }]
                    }
                }
            }
        };
        this.$item = {
            $category: "section",
            $title: localize.cfa_section_title,
            $layout: {
                $items: [{
                    $bind: "$format",
                    $isEditMode: true,
                    $labelWidth: "auto",
                    $skin: "s-aw-field"
                }, {
                    $bind: "$choiceLayout",
                    $format: "$combo",
                    $isEditMode: true,
                    $labelWidth: "auto",
                    $skin: "s-aw-field"
                }]
            }
        };
        Article.prototype.loadBox.call(this, initData);
    },
    dispose: function(){
        if (this.designedCombo && this.designedCombo.builder) {
            this.designedCombo.builder.onAuthoring = false;
            this.designedCombo.builder.toggleList(false);
        }
        delete this.designedCombo;
        Article.prototype.dispose.call(this);
    }
});
