// syracuse-ui/lib/field/choice/comboFormat.js
"use strict";
exports.$defaultChoiceLayout = "1";

function _setIcon(field, dom, $value) {
	var $path = syra_site.$iconPath + (field.$item.$icon && field.$item.$icon.$path || "");
	dom.style.backgroundImage = $value ? "url('" + $path + $value + ".png')" : "";
}

function _createIcon(field, $value, css) {
	var img = document.createElement("div");
	img.className = css;
	_setIcon(field, img, $value);
	return img;
}

function _setItemText(field, $enum, filter) {
	if (filter) {
		if ((new RegExp("^" + _escapeRegex(filter), "i")).test($enum.$title)) {
			var className = field.currentValue && $enum.$value == field.currentValue ? 's-auto-complete-filter selected' : 's-auto-complete-filter';
			return $enum.$title.replace(new RegExp("(" + _escapeRegex(filter) + ")", "i"), "<strong class='" + className + "'>$1</strong>");
		}
	} else {
		return $enum.$title;
	}
}

function _selectItem(field, $index, select) {
	if (field.$selectedEnum) {
		delete field.$selectedEnum;
	}
	if (select) {
		field.$selectedEnum = field.$enum[$index];
		field.$selectedEnum && _selectChoice(field, field.$selectedEnum.$index = $index);
	}
}

function _selectChoice(field, selectedIndex) {
	if (field._choices) {
		for (var ii = 0, jj = field._choices.length; ii < jj; ii++) {
			var choice = field._choices[ii];
			if (choice.link) {
				syra_site.dom.toggleClass(choice.link, "s-selected", choice.$index == selectedIndex);
			}
		}
	}
}

function _addChoiceItem(field, index, filter, cols) {
	var choice = {
		$index: index
	};
	var $enum = field.$enum[index];
	var html = _setItemText(field, $enum, filter);
	if (html != null) {
		if (html == "") {
			html = "&nbsp;";
		}
		if (field.$selectedEnum && index == field.$selectedEnum.$index) {
			field._$selectedListIndex = field._choices.length;
		}
		var css = field.$skinListItem;
		if (field.$item.$css) {
			css += " " + field.$item.$css;
		}
		choice.link = syra_menus.addIconButton("", css, "onChoiceClick");
		choice.link.syraChoice = index;
		choice.link.syraOnChoiceItemClick = true;
		if (field.$item.$icon) {
			choice.link.className += " s-icon";
			choice.link.appendChild(_createIcon(field, $enum.$value, field.$skinListItem + "-icon"));

			if (field.$item.$icon.$mode == 'iconText') {
				var div = document.createElement("div");
				div.className = field.$skinListItem + "-iconText";
				div.innerHTML = html;
				choice.link.appendChild(div);
			} else {
				if (field.$item.$icon.$mode != "icon") {
					choice.link.innerHTML = html;
				} else {
					syra_menus.setButtonTitle(choice.link, html);
				}
			}

		} else {
			if (field.$item.$cssMode) {
				choice.link.className += " s-cssmode " + field.$item.$cssMode.$css;
				var div = document.createElement("div");
				div.syraCss = field.$skinListItem + "-cssmode " + field.$item.$cssMode.$css;
				_setCssValue(field, div, $enum.$value);
				choice.link.appendChild(div);
			} else {
				choice.link.innerHTML = html;
			}
		}
		if (field.currentValue == $enum.$value) {
			choice.link.className += " s-selected";
		}
		if (!cols) {
			choice.link.style.display = "inline-block";
			field._list.appendChild(choice.link);
		} else {
			cols[field.curCol++].appendChild(choice.link);
			if (field.curCol >= cols.length) {
				field.curCol = 0;
			}
		}
		field._choices.push(choice);
	}
}



exports.renderLayout = function(field, filter) {
	field._table = null;
	field._choices = [];
	if (field._list) {
		syra_site.dom.empty(field._list);
		var $choiceLayout = field.$item.$choiceLayout;
		var cols;
		if ($choiceLayout != "row") {
			$choiceLayout = parseInt($choiceLayout, 10);
			cols = [];
			field._table = document.createElement("div");
			field._table.className = "s-choice-list-table";
			field._list.appendChild(field._table);
			for (var ii = 0, colCount = $choiceLayout; ii < colCount; ii++) {
				var col = document.createElement("div");
				col.className = "s-choice-list-col";
				cols[ii] = field._table.appendChild(col);
			}
		}
		field._$selectedListIndex = -1;
		if (filter === "") {
			filter = null;
		}
		field.curCol = 0;
		for (var ii = 0, jj = field.$enum.length; ii < jj; ii++) {
			_addChoiceItem(field, ii, filter, cols);
		}
		if (filter && field._choices.length == 0) {
			var noMatch = document.createElement("div");
			noMatch.className = "s-choice-no-match";
			noMatch.textContent = syra_local.fieldChoiceNoMatch;
			field._list.appendChild(noMatch);
		}
		_setMinWidth(field);
	} else {
		for (var ii = 0, jj = field.$enum.length; ii < jj; ii++) {
			var $enum = field.$enum[ii];
			if (filter) {
				if ((new RegExp("^" + _escapeRegex(filter), "i")).test($enum.$title)) {
					var className = field.currentValue && $enum.$value == field.currentValue ? 's-auto-complete-filter selected' : 's-auto-complete-filter';
					field._choices.push({
						$index: ii
					});
				}
			}
		}
	}
};


function _setCssValue(field, dom, $value) {
	dom.className = dom.syraCss;
	if ($value != undefined) {
		if (field.$item.$cssMode.$isColor) {
			dom.style.backgroundColor = $value;
		} else {
			dom.className += " " + "s-choice-css-" + $value;
		}
	}
}

function _endInputFilter(field) {
	clearTimeout(field._autoCompleteTimer);
	var closed = _toggleList(field, false);
	delete field.autoCompleteFilter;
	return closed;
}

function _onChoiceItemClick(field, btn) {
	var index = btn.$index;
	if (index === undefined) {
		index = btn.syraChoice;
	}
	delete field.autoCompleteFilter;
	if (index !== undefined && (!field.$isDisabled && (!field.$isReadOnly || field.$field.$isAutoComplete === false))) {
		_selectItem(field, index, true);
		field.setInputValue(field.$selectedEnum.$title);
		$(field.input).change();
		if (field.iconValue) {
			_setIcon(field, field.iconValue, field.$selectedEnum ? field.$selectedEnum.$value : null);
		}
		if (field.cssValue) {
			_setCssValue(field, field.cssValue, field.$selectedEnum ? field.$selectedEnum.$value : null);
		}
		_toggleList(field, false);
		if (field.page && field.page.focusField != field) {
			field.focus();
		}
	}
}

function _filterInput(field, charCode, isSelectDisabled) {
	clearTimeout(field._autoCompleteTimer);
	if (field.$field.$isAutoComplete === false) {
		field.autoCompleteFilter = (field.autoCompleteFilter || "") + charCode;
	}
	field._autoCompleteTimer = setTimeout(function() {
		var filter;
		if (field.autoCompleteFilter !== undefined) {
			filter = field.autoCompleteFilter;
		} else {
			filter = field.input.value;
		}
		exports.renderLayout(field, filter);
		if (!isSelectDisabled && field._choices.length == 1) {
			_onChoiceItemClick(field, field._choices[0]);
			_toggleList(field, false);
		} else {
			_toggleList(field, true, filter);
		}

	}, 100);
}

function _applyShortCut(field, shortcuts, event) {
	if (!field.$isDisabled && !field.$isReadOnly) {
		if (shortcuts.backspace) {
			if (field.autoCompleteFilter) {
				field.autoCompleteFilter = field.autoCompleteFilter.slice(0, field.autoCompleteFilter.length - 1);
			}
			_filterInput(field, "", true);
			if (field.$field.$isAutoComplete === false) {
				event.preventDefault();
				event.stopPropagation();
			}
			return true;
		}
		if (shortcuts.down || shortcuts.up) {
			clearTimeout(field._autoCompleteTimer);
			var isOpened = field.popupPicker && field.popupPicker.isOpened;
			if (!isOpened && shortcuts.up) {
				return false;
			}
			if (isOpened && event.shiftKey && shortcuts.up) {
				field.applyEscape();
				event.preventDefault();
				event.stopPropagation();
				return true;
			}
			_toggleList(field, true);
			var current = field._$selectedListIndex !== undefined ? field._$selectedListIndex : -1;
			if (isOpened) {
				if (!field._choices.length) {
					_toggleList(field, false);
					event.preventDefault();
					event.stopPropagation();
					return true;
				}
				if (shortcuts.down) {
					current++;
					if (current >= field._choices.length) {
						current = 0;
					}
				} else {
					current--;
					if (current < 0) {
						current = field._choices.length - 1;
					}
				}
			}
			if (!field._choices.length)
				return;
			field._$selectedListIndex = Math.max(current, 0);
			_selectChoice(field, field._choices[field._$selectedListIndex].$index);
			field.input.value = field.$enum[field._choices[field._$selectedListIndex].$index].$title;
			_selectItem(field, field._choices[field._$selectedListIndex].$index, true);
			delete field.autoCompleteFilter;
			event.preventDefault();
			event.stopPropagation();
			return true;
		}
		if (shortcuts.enter) {
			clearTimeout(field._autoCompleteTimer);
			var filter;
			if (field.autoCompleteFilter !== undefined) {
				filter = field.autoCompleteFilter;
				delete field.autoCompleteFilter;
			} else {
				filter = field.input.value;
			}
			exports.renderLayout(field, filter);
			if (field._choices.length > 0) {
				_onChoiceItemClick(field, field._choices[0]);
				if (field._choices.length == 1) {
					return _toggleList(field, false);
				}
			}
			return false;
		}
	}
}


function _toggleList(field, show, filter) {
	if (!field.popupPicker && show !== false) {
		if (field.page.focusField != field) {
			field.focus();
		}
		field.domItem.id = field.id;
		field._list = document.createElement("div");
		field._list.className = "s-choice-list-popup";
		if (field.$item.$css) {
			field._list.className += " " + field.$item.$css;
		}
		field._list.syraItem = field.id;
		field.articleParent.setArticleId(field._list);
		field._list.style.display = "none";
		exports.renderLayout(field, filter);
		if (field.$selectedEnum) {
			_selectItem(field, field.$selectedEnum.$index, true);
		}
		field.popupPicker = syra_site.dialogManager.openPopup(field.boxParent, {
			content: field,
			autoCloseBoundary: field.domItem,
			slot: field._list,
			picker: field.layoutSlot,
			position: {
				my: "left top",
				at: "left bottom",
				of: $(field._dataValue)
			},
			onresize: function(dialog) {
				if (field.input) {
					var parentRect = field.input.parentNode.getBoundingClientRect();
					var dialogRect = dialog.dialogSlot.getBoundingClientRect();
					var minWidth = parentRect.width + "px";
					field._list.style.minWidth = minWidth;
					var diff = field._list.scrollHeight - field._list.clientHeight;
					if (diff) {
						dialog.dialogSlot.style.height = Math.min(dialog._maxHeight, dialogRect.height + diff) + "px";
					}
					if (field._table) {
						field._table.style.width = field._list.clientWidth + "px";
					}
				}
			},
			onClose: function() {
				setTimeout(function() {
					if (field && field.input) { //need test on field input for list pagging field 
						field.input.value = field.$selectedEnum ? field.$selectedEnum.$title : "";
						syra_site.dom.removeChild(field._list);
					}
					field.popupPicker = field._list = null;
				}, 200);
			}
		});
		return true;
	} else {
		if (field.popupPicker) {
			if (show) {
				_setMinWidth(field);
				field.popupPicker.resizeDialog();
				return false;
			} else {
				field.popupPicker.close();
				return true;
			}
		}
	}
}

function _escapeRegex(string) {
	return string.replace(/[\^\$\.\*\+\-\?\=\!\:\|\\\/\(\)\[\]\{\}\,]/g, '\\$&');
}

function _selectText(input, start, end) {
	if (input.createTextRange) {
		var selRange = input.createTextRange();
		selRange.collapse(true);
		selRange.moveStart('character', start);
		selRange.moveEnd('character', end - start);
		selRange.select();
	} else
	if (input.selectionStart) {
		input.selectionStart = start;
		input.selectionEnd = end;
	}
}


function _fillSelectedEnum(field, $selectedEnum) {
	var $title = $selectedEnum ? $selectedEnum.$title : "";
	if (field.$isEditMode) {
		field.setInputValue($title);
		if (field.iconValue) {
			_setIcon(field, field.iconValue, $selectedEnum ? $selectedEnum.$value : null);
		}
		if (field.cssValue) {
			_setCssValue(field, field.cssValue, $selectedEnum ? $selectedEnum.$value : null);
		}
	} else {
		if ($title && !field.$isDetailLinkDisabled && field.$menus && field.$menus.$details) {
			field.appendDetailLink($selectedEnum.$value, field.$menus.$details);
		} else {
			field._dataValue.textContent = field.displayText = $title;
		}
	}
}

exports.dispose = function(field) {
	if (field) {
		field._list = field._table = field.iconValue = field.cssValue = null;
		field.getDataValue = field.onInputValueChanged = field.applyEscape = field.applyShortCut = field.onChoiceClick = field.validateKeyPress = null;
	}
};

exports.setEnum = function(field, $enum) {
	field.$enum = $enum;
	field.localizeEnum();
	field._choices = [];
};

exports.render = function(field) {
	field.getDataValue = function() {
		if (this.$isEditMode && this.$field.$isAutoComplete !== false) {
			if (this.getInputValue() == "") {
				this.$selectedEnum = null;
			}
		}
		return this.$selectedEnum ? this.$selectedEnum.$value : null;
	};
	field.onInputValueChanged = function(event) {
		_fillSelectedEnum(this, this.findEnum(this.getDataValue()));
		return true;
	};
	field.applyEscape = function(shortcuts, event) {
		return _endInputFilter(this);
	};
	field.applyShortCut = function(shortcuts, event) {
		return _applyShortCut(this, shortcuts, event);
	};
	field.onChoiceClick = function(event, btn) {
		if (btn.syraOnChoiceItemClick) {
			_onChoiceItemClick(this, btn);
		} else {
			_toggleList(this);
		}
	};
	field.validateKeyPress = function(charCode, event) {
		if (event.keyCode != 13) {
			_filterInput(this, charCode);
		}
		return true;
	};
	//field.$field.$isAutoComplete = false;
	field.pickerCount = 1;
	var $value = field.$field.$value;
	delete field.$item.$isAutoSizeDisabled;
	exports.setEnum(field, field.$enum = $value ? $value.$enum : {});
	this.$skinListItem = "s-choice-list-item";
	if (field.$isEditMode) {
		if (field.$item.$contentEditable) {
			field.input = document.createElement("div");
			field.input.setAttribute("contenteditable", "true");
		} else {
			field.input = document.createElement("input");
			field.input.setAttribute("type", "text");
		}
		field.input.syraIsInput = true;
		field.input.syraOnClick = "onInputFieldClick";
		field._dataValue.className += " s-combo";
		syra_site.setSpecificAttributes(field.input);

		if (field.$field.$isAutoComplete === false) {
			field.input.setAttribute("readonly", "readonly");
		}
		field.input.className = field.$skinInput;
		if (field.$item.$contentEditable) {
			field.input.className += " s-contenteditable";
		}
		field._dataValue.appendChild(field.input);
		if (!field.mnPickers.choice) {
			syra_menus.addFieldPicker(field, "choice", syra_local.fieldOpenChoiceList, "SHIFT DOWN");
		}
		if (field.$item.$icon) {
			field.mnPickers.choice.className += " s-icon";
			field.iconValue = _createIcon(field, field.currentValue, field.$skin + "-choice-icon");
			$(field._dataValue).prepend(field.iconValue);
			field.iconValue.top = "2px";
			field.iconValue.left = "2px";
			field.input.style.paddingRight = (field._boxPickerPaddingRight - 20) + "px";
			if (field.$item.$icon.$inputMode == "icon") {
				field.input.style.width = "0px";
				field.input.style.visibility = "hidden";
				field.domItem.className += " s-choice-icon-mode";
			}
		}
		if (field.$item.$cssMode) {
			field.mnPickers.choice.className += " s-cssmode " + field.$item.$cssMode.$css;
			field.cssValue = document.createElement("div");
			field.cssValue.syraCss = field.$skin + "-choice-cssmode " + field.$item.$cssMode.$css;
			_setCssValue(field, field.cssValue, field.currentValue);
			field.cssValue.syraCss = field.$skin + "-choice-cssmode " + field.$item.$cssMode.$css;
			field._dataValue.insertBefore(field.cssValue, field._dataValue.firstChild);
			field.input.className += " " + field.$skin + "-choice-cssmode-input";
			if (field.$item.$cssMode.$inputMode == "css") {
				field.input.style.display = "none";
			}
		}
		if (field.$item.$css) {
			field.input.className += " " + field.$item.$css;
		}
	}
};

exports.setDataValue = function(field, value) {
	field.currentValue = value;
	_fillSelectedEnum(field, field.$selectedEnum = field.findEnum(value));
};

function _setMinWidth(field) {
	field._list.style.minWidth = field.input.parentNode.clientWidth + "px";
}
