// syracuse-ui/lib/field/choice/radiosFormat.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');

function RadiosFormat(){
    this.$defaultChoiceLayout = "row";
}

exports.RadiosFormat = helpers.defineClass(RadiosFormat, null, {
    onInputKeyup: function($$input, event){
        return true;
    },
    setDataValue: function(value){
        this.field.currentValue = value;
        this._$selectedEnum = this.field.findEnum(value);
        if (this.field.$isEditMode) {
            if (this._$selectedEnum) {
                $(this.field.fieldValue).find("input[value='" + this._$selectedEnum.$index + "']").attr("checked", true);
            }
        }
        else {
            this.field.fieldValue.textContent = this._$selectedEnum ? this._$selectedEnum.$title : "";
        }
    },
    getDataValue: function(){
        var index = this._$$fieldset.find("input").filter(":checked").val();
        return (index == null) ? null : this.field.$enum[index].$value;
    },
    setEnum: function($enum){
        this.field.$enum = $enum;
        if (this._choices) {
            this.renderLayout();
        }
    },
    render: function(){
        this.field.$enum = this.field.getDataType().$enum;
        if (this.field.$isEditMode) {
            var div = document.createElement("fieldset");
            div.className = this.$skin = this.field.$skin + "-choice-rd";
            this._$$fieldset = $(this.field.fieldValue.appendChild(div));
            this.renderLayout();
        }
        else {
            //this.field.$$fieldValue.addClass("");
        }
    },
    renderLayout: function(){
        var self = this;
        self.field.$isLayoutContentSizeDisabled = true;
        self.releaseMode();
        self._choices = [];
        self._$$fieldset.empty();
        var isColLayout = self.field.$item.$choiceLayout != "row";
        var $$rows, curRow = 0, colCount, curCol = 0;
        if (isColLayout) {
            $$rows = [];
            colCount = self.field.$item.$choiceLayout;
            for (var ii = 0, max = Math.ceil(self.field.$enum.length / colCount); ii < max; ii++) {
                var div = document.createElement("div");
                div.className = "s-field-choice-rd-row";
                $$rows[ii] = $(div).appendTo(self._$$fieldset);
            }
        }
        var $skinItem = self.$skin + "-item";
        var $groupName = helpers.uuid.generate();
        self.field.$enum.forEach(function($enum, index){
            var id = $groupName + index;
            var input = document.createElement("input");
            input.type = "radio";
            var $$input = $(input).val(index);
            input.className = self.$skin + "-radio";
            input.setAttribute("id", id);
            input.setAttribute("name", $groupName);
            input.setAttribute("data-s-field", self.field.id);
            
            var domTitle = document.createElement("label");
            domTitle.className = self.$skin + "-title";
            domTitle.setAttribute("for", id);
            domTitle.textContent = $enum.$title;
            
            var domItem = document.createElement("div");
            domItem.className = $skinItem;
            domItem.appendChild(input);
            var $$item = $(domItem);
            if (isColLayout) {
                $$rows[curRow].append($$item);
                if (++curCol == colCount) {
                    curRow++;
                    curCol = 0;
                }
            }
            else {
                self._$$fieldset.append($$item);
            }
            self._choices.push({
                $index: index,
                $$item: $$item
            });
            if (self.field.$item.$icon) {
                var image = self.$icon.$path + $enum.$value + ".png";
                var img = document.createElement("div");
                img.className = $skinItem + "-icon";
                img.style.backgroundImage = "url('" + image + "')";
                domItem.appendChild(img);
            }
            domItem.appendChild(domTitle);
        });
        if (isColLayout && curCol > 0) {
            while (curCol < colCount) {
                var html = document.createElement("div");
                html.className = self.$skin + "-item";
                $$rows[curRow].append($(html));
                curCol++;
            }
        }
    },
    releaseMode: function(){
        delete this._choices;
    },
    dispose: function(){
        this.releaseMode(true);
        delete this.field;
    }
});
