// syracuse-tablet/html/js/helpers/native/wrapperTest/nativeTestPenInput.js
"use strict";

var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("nativePenInput");

var _TestNativeWrapper = function() {

	this.getPenTextInputSupported = function() {
		return true;
	};
	this.getAnnotatedImageSupported = function() {
		return true;
	};
	this.getSignatureImageSupported = function() {
		return false;
	};
	this.getPenTextInput = function(context) {
		context = context ? JSON.parse(context) : {};
		var text = window.prompt(context.Title || "Enter the text", context.Value || "");
		var data = {
			Action: text == null || text.trim().length === 0 ? "cancelled" : "ok",
			//Returns an array of words
			Matches: text ? text.split(' ') : null
		};
		setTimeout(function() {
			window.smPenInputJS.penInputTextFinished(data);
		}, 50);
	};
	this.getAnnotatedImage = function(options) {
		var img = "";
		setTimeout(function() {
			var data = {
				Action: "ok",
				ImageData: "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAUCAIAAAAC64paAAAACXBIWXMAAAsSAAALEgHS3X78AAAAGklEQVR42mOYacxANmIY1TyqeVTzqOaB1QwApdw+0KHEWp0AAAAASUVORK5CYII="
			};
			log && log("Annotated Image GET ");
			window.smPenInputJS.penAnnotateImageFinished(data);
		}, 100);
	};

	this.getSignatureImageSupported = function(options) {
		return true;
	};
	this.getSignatureImage = function(options) {
		var img = "";
		setTimeout(function() {
			var data = {
				Action: "ok",
				ImageData: "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAUCAIAAAAC64paAAAACXBIWXMAAAsSAAALEgHS3X78AAAAGklEQVR42mOYacxANmIY1TyqeVTzqOaB1QwApdw+0KHEWp0AAAAASUVORK5CYII="
			};
			log && log("Signature Image GET ");
			window.smPenInputJS.penSignatureImageFinished(data);
		}, 100);
	};

};

exports.create = function() {
	return new _TestNativeWrapper();
};
