// syracuse-tablet/html/js/wc/pages/sdata/pageHome.js
"use strict";

var Base = require('syracuse-tablet/html/js/wc/pages/sdata/pageSdataDashboard').Page;
var utils = require('syracuse-tablet/html/js/wc/helpers/utils');
var locale = require('syracuse-tablet/html/js/wc/helpers/locale');
var HomeController = require("syracuse-tablet/html/js/wc/controllers/homeController").HomeController;

/**
 *
 */
exports.Page = utils.defineClass(
	function PageHome(pageName, options) {
		Base.call(this, pageName, options);
		this.actionAdapter.showDesignPage = false;
	},
	Base, {
		destroy: function() {
			Base.prototype.destroy.call(this);
		},
		getPageTitle: function() {
			return locale.text("global.pageTitle.home");
		},
		isRootPage: function() {
			return true;
		},
		_ensureController: function() {
			this.controller = new HomeController(this.pageData.dataset, null, this.pageData.isWorkingCopy);
			// Will raise notifications or do confirmations
			this.controller.setUIAdapter(this);
			this.controller.setNativeCapabilities(this.options)
		},
	}
);
