// syracuse-tablet/html/js/wc/pages/sdata/pageHome.js
"use strict";

var Base = require('syracuse-tablet/html/js/wc/pages/sdata/pageSdataDashboard').Page;
var utils = require('syracuse-tablet/html/js/wc/helpers/utils');
var locale = require('syracuse-tablet/html/js/wc/helpers/locale');
var HomeController = require("syracuse-tablet/html/js/wc/controllers/homeController").HomeController;
var native = require('syracuse-tablet/html/js/wc/helpers/native/native');
var wpHelpers = require('syracuse-tablet/html/js/wc/sdata/wpHelpers');

/**
 *
 */
exports.Page = utils.defineClass(
	function PageHome(pageData, options) {
		Base.call(this, pageData, options);
		this.actionAdapter.showDesignPage = false;
		this.actionAdapter.showPageShare = false;
		this.actionAdapter.showMailTo = false;
		this.actionAdapter.showPinPage = false;
		this.actionAdapter.showHome = false;
		this.actionAdapter.showPageConfig = true;
		this.nativeVoiceCommands = native.getModule("voiceCommands");
	},
	Base, {
		destroy: function() {
			Base.prototype.destroy.call(this);
		},
		getPageTitle: function() {
			return locale.text("global.pageTitle.home");
		},
		isRootPage: function() {
			return true;
		},
		_ensureController: function() {
			this.controller = new HomeController(this.pageData.dataset, null, this.pageData.isWorkingCopy);
			// Will raise notifications or do confirmations
			this.controller.setUIAdapter(this);
			this.controller.setNativeCapabilities(this.options)
		},

		afterRender: function() {
			Base.prototype.afterRender.call(this);
			this.updateVoiceCommands();
		},

		updateVoiceCommands: function() {
			var self = this;
			if (self.nativeVoiceCommands) {
				self.getVoiceCommands().then(function(commands) {
					self.nativeVoiceCommands.updateVoiceCommands(commands);
				});
			}
		},
		_actPageConfig: function() {
			alert("Edit button called by button in welcome page");
			// TODO: https://jira-sage.valiantyscloud.net/browse/X3-17956
		},
		/**
		 * List of voice commands that can be called by user
		 * will be stored in native api and open the called page if the title matches
		 * 
		 * Promise resolving with:
		 * [
		 * {
		 *   title: "My Customers",
		 *   page: {
		 *     $url: ...
		 *     $method: ...
		 *   }
		 * ]
		 */
		getVoiceCommands: function() {
			return $.smResolve(wpHelpers.welcomePageToCommands(this.pageData));
		}
	}
);
