// syracuse-tablet/html/js/wc/pages/sdata/pageSdataChildPage.js
"use strict";

var Base = require('syracuse-tablet/html/js/wc/pages/sdata/pageSdata').Page;
var modules = require('syracuse-tablet/html/js/wc/common/modules');
var articleParser = require("syracuse-tablet/html/js/wc/controls/articleParser");
var SdataController = require("syracuse-tablet/html/js/wc/controllers/sdataController").SdataController;
var utils = require('syracuse-tablet/html/js/wc/helpers/utils');

/**
 * Page handling a query
 * We should add filters and paging
 * 
 */
exports.Page = utils.defineClass(
	function PageSdataChildPage(pageData, options) {
		this.parentController = modules.get("controller").findController(options.childContext.controllerId);
		pageData = this._createPageData();

		Base.call(this, pageData, options);
	},
	Base, {
		destroy: function() {
			Base.prototype.destroy.call(this);
		},
		_createPageData: function() {
			var dataset = this.parentController.dataset;
			return {
				dataset: dataset,
				prototype: dataset.prototype,
				article: articleParser.createDefaultArticle(dataset.prototype)
			};
		},
		_ensureController: function() {
			this.controller = new SdataController(this.pageData.dataset, this.parentController, this.pageData.isWorkingCopy);
			// Will raise notifications or do confirmations
			this.controller.setUIAdapter(this);
		}
	}
);
