// syracuse-tablet/html/js/wc/pages/sdata/pageSdataChartDetail.js
"use strict";

var Base = require('syracuse-tablet/html/js/wc/pages/sdata/pageSdata').Page;
var articleParser = require("syracuse-tablet/html/js/wc/controls/articleParser");
var siteLayout = require('syracuse-tablet/html/js/wc/ui/siteLayout');

var utils = require('syracuse-tablet/html/js/wc/helpers/utils');

/**
 * Page handling a query
 * We should add filters and paging
 * 
 */
exports.Page = utils.defineClass(
	function PageSdataChartDetail(pageData, options) {
		pageData.page.$article = this._getDefaultArticle(pageData.prototype)
		Base.call(this, pageData, options);
		this.controller.level = options.level || 0;
		this.pageName += "_" + this.controller.level;
	},
	Base, {
		destroy: function() {
			Base.prototype.destroy.call(this);
		},
		scrollAllowed: function() {
			return true;
		},
		_getDefaultArticle: function(prototype) {
			//TODO : get article depending on level and page settings
			return articleParser.createDefaultChartDetailArticle(prototype);
		},
		_getDefaultAttachedControls: function() {
			var device = siteLayout.getDeviceType();
			if (device === "smartphone") {
				return ["header", "chartDetailToolBar"];
			} else {
				return ["header", "breadcrumbs", "chartDetailToolBar", "footer"];
			}
		},
		_setDisplay: function() {
			this.authUpdateLayout(this.getArticle())
		},
		actToolbarSetDisplay: function(opts) {
			this._setChartDetailPrefs(opts);
			this._setDisplay();
			this.writePageSettings()
				.then(function() {
					// just to document it's a promise
				});
		},
		_setChartDetailPrefs: function(opts) {
			var settings = this.getPageSettings();
			settings.chartDetail = settings.chartDetail || {};
			settings.chartDetail[0 /*this.state.level*/ ] = opts;
			settings.isDirty = true;
		},
		isChartDetail: function() {
			return true
		},
		getAuthoringName: function() {
			var pageAuthName = Base.prototype.getAuthoringName.call(this);
			pageAuthName += "_" + this.controller.level;
			console.log("pageAuthName ---->" + pageAuthName)
			return pageAuthName;
		}
	}

);
