// syracuse-tablet/html/js/wc/pages/sdata/pageSdataDashboard.js
"use strict";

var Base = require('syracuse-tablet/html/js/wc/pages/sdata/pageSdata').Page;
var notifications = require('syracuse-tablet/html/js/wc/helpers/notifications');

var utils = require('syracuse-tablet/html/js/wc/helpers/utils');
var protoHelpers = require('syracuse-tablet/html/js/wc/sdata/protocolHelpers');

/**
 * Page handling a dashboard
 * 
 * options: { // Additional options for dashboard page. These add up to what is possible for pageBase
 *   dashboardParameters: {
 *     CODE: {
 *       value: "10"
 *     }
 *   }
 * }
 */
exports.Page = utils.defineClass(
	function PageSdataDashboard(pageData, options) {
		Base.call(this, pageData, options);
		this._nbVignettesLoaded = 0;
		// Default, will be updaten on every loaded vignette
		this._dataFreshness = {
			dateTime: new Date().getTime(),
			level: "fresh"
		};
		this._savedVignetteContext = null;
	},
	Base, {
		destroy: function() {
			Base.prototype.destroy.call(this);
			this._savedVignetteContext = null;
			this._vignettes = null;
		},

		build: function() {
			this._vignettes = null;
			return Base.prototype.build.call(this);
		},

		afterRender: function() {
			Base.prototype.afterRender.call(this);
			this._nbVignettesLoaded = 0;
		},

		getPageTitle: function() {
			// Give by dashboard value
			var ttl = this.controller.dataset.getValue("title");
			if (!ttl) {
				// Compliant with stateless dashboard (client test app)
				ttl = ttl = this.controller.dataset.getValue("$title");
			}
			if (!ttl) {
				// Given by prototype
				ttl = Base.prototype.getPageTitle.call(this);
			}
			return ttl
		},

		onVignetteLoaded: function(vignette, success) {
			if (success) {
				this._updateDataFreshness(vignette);
			}
			this._nbVignettesLoaded++;
			var vignettes = this._getVignettes();
			if (vignettes.length == 0) {
				this.afterAllVignettesLoaded();
			} else if (this._nbVignettesLoaded >= vignettes.length) {
				var self = this;
				setTimeout(function() {
					// Once all vignettes have been loaded (eg: stack hub scroller height depends on vignette content)
					vignettes.forEach(function(v) {
						v.afterAllVignettesLoaded();
					})
					self.afterAllVignettesLoaded();
				})
			}
		},
		afterAllVignettesLoaded: function() {
			this.onContentChanged();
			notifications.publish(["sm.all.vignettes.loaded"], this);
		},
		/*
		 * Called whenever a vignette is loaded to update the overall freshness of the dashboard
		 */
		_updateDataFreshness: function(vignette) {
			var childPage = vignette && vignette.getTopPage();
			var dataset = childPage && childPage.controller && childPage.controller.dataset;
			if (dataset) {
				var freshness = dataset.getDataFreshness();
				if (this._dataFreshness.dateTime > freshness.dateTime) {
					this._dataFreshness.dateTime = freshness.dateTime;
					this._dataFreshness.level = freshness.level;
					notifications.publish("sm.data.freshness.change", this._dataFreshness);
				}
			}
		},
		/**
		 * Returns name to use for saving authoring
		 * syracuse.collaboration.testHomeDashboard.$mobileDashboard
		 */
		getAuthoringName: function() {
			var $url = this.pageData.$url;
			var key = this.pageData.dataset.json.dashboardName;
			var ep = protoHelpers.getEndpointFromUrl($url);
			ep = ep.split(".").slice(0, 2).join(".");
			return ep + "." + key + ".$mobileDashboard";
		},
		/**
		 * typeDashboard = hub / stack / undefined
		 */
		isDashboard: function(typeDashboard) {
			return typeDashboard == null ? true : (this.rootLayout ? this.rootLayout.$type === typeDashboard : true);
		},
		/**
		 * Scroll handled by hub layout
		 */
		scrollAllowed: function() {
			return !this.isDashboard("hub");
		},
		getGestureMgr: function() {
			return this.isDashboard("hub") && this.rootLayout ? this.rootLayout.getGestureMgr() : Base.prototype.getGestureMgr.call(this);
		},
		savedCtxCreate: function(opts) {
			if (this.destroyed) {
				return;
			}
			var ctx = Base.prototype.savedCtxCreate.call(this, opts);
			ctx.vignettes = [];
			var p;
			this._getVignettes().forEach(function(v) {
				ctx.vignettes.push(v.getTopPage() ? v.getTopPage().savedCtxCreate(opts) : null);
			});
			if (this.isDashboard("hub") && this.rootLayout) {
				ctx.hubCtx = this.rootLayout.savedCtxCreate(opts);
			}
			return ctx;
		},
		savedCtxRestore: function(ctx) {
			if (this.destroyed) {
				return;
			}
			Base.prototype.savedCtxRestore.call(this, ctx);
			// store vignettes context because it is removed after calling savedCtxRestore
			// Vignettes are loaded asynch so they will need the context later
			this._savedVignetteContext = ctx.vignettes;
			if (ctx.hubCtx && this.isDashboard("hub")) {
				// Restore dashboard context
				this.rootLayout.savedCtxRestore(ctx.hubCtx);
			}
		},
		getSavedCtxVignette: function(v) {
			if (v == null) {
				return null;
			}
			if (this._savedVignetteContext == null) {
				// Check page context
				// Sometimes we want to read the context before savedCtxRestore (see searchArray)
				this._savedVignetteContext = (this.options.savedCtx && this.options.savedCtx.vignettes) || [];
			}
			var idx = this._getVignettes().indexOf(v);
			if (idx == -1 || idx > this._savedVignetteContext.length - 1) {
				return null;
			}
			return this._savedVignetteContext[idx];
		},

		_getVignettes: function() {
			if (this._vignettes == null) {
				this._vignettes = this.controller.getControlsByType("application/x-vignette") || [];
			}
			return this._vignettes;
		}
	}
);
