// syracuse-tablet/html/js/wc/pages/sdata/pageSdataCardDetail.js
"use strict";

var Base = require('syracuse-tablet/html/js/wc/pages/sdata/pageSdataChildPage').Page;
var utils = require('syracuse-tablet/html/js/wc/helpers/utils');
var locale = require('syracuse-tablet/html/js/wc/helpers/locale');
var articleParser = require("syracuse-tablet/html/js/wc/controls/articleParser");

/**
 * Page handling a query
 * We should add filters and paging
 * 
 */
exports.Page = utils.defineClass(
	function PageSdataCardDetail(pageData, options) {
		Base.call(this, pageData, options);

		this.actionAdapter.showAccept = false;
		// Cancel deactivated since backup and revert is not fully implemented yet
		this.actionAdapter.showCancel = false; // true;

		// No row detail authoring at the moment
		this.actionAdapter.showDesignPage = false;
		this.actionAdapter.showPinPage = false;
		this.actionAdapter.showHome = false;

	},
	Base, {
		destroy: function() {
			Base.prototype.destroy.call(this);
		},
		_createPageData: function(pageData) {
			var childPageData = this.options.childContext.cardDetailPageData;
			if (!childPageData.page.$article) {
				childPageData.page.$article = this._getArticle(childPageData.prototype);
			}
			return childPageData;
		},
		_getArticle: function(prototype) {
			return articleParser.createDefaultArticle(prototype);
		},
		getCardArrayControl: function() {
			var parentPageController = this.controller.findController(this.options.childContext.controllerId);
			if (parentPageController) {
				var arrayControl = parentPageController.getControl(this.options.childContext.arrayControlId);
				if (arrayControl) {
					return arrayControl;
				}
			}
		},
		getCardArticle: function() {
			var cardArticle = this.pageData.page.$article;
			return cardArticle;
		},
		getPageTitle: function() {
			return locale.text("auth.designCardLayout");
		},
		isCardDetail: function() {
			return true;
		}
	}
);
