// syracuse-tablet/html/js/wc/pages/sdata/pageSdataQuery.js
"use strict";

var Base = require('syracuse-tablet/html/js/wc/pages/sdata/pageSdata').Page;
var modules = require('syracuse-tablet/html/js/wc/common/modules');
var utils = require('syracuse-tablet/html/js/wc/helpers/utils');

/**
 * Page handling a query
 * We should add filters and paging
 * 
 */
exports.Page = utils.defineClass(
	function PageSdataQuery(pageData, options) {
		Base.call(this, pageData, options);
		this.mutiSelectionEnabled = false;
		this._multiselPanel = null;
	},
	Base, {
		destroy: function() {
			Base.prototype.destroy.call(this);
			this._multiselPanel = null;
			this._multiSelArray = null;
		},
		beforeOnSdataActionClicked: function(actionName, controller) {
			var self = this;
			return $.smResolve().then(function() {
				if (actionName === "multiSelection") {
					self.multiSelToggleStatus();
					return false;
				}
				return true;
			})
		},
		multiSelToggleStatus: function() {
			if (!this.attachedControlsMgr) {
				return;
			}
			var regularPanel;
			var panelInfo;
			if (modules.get("siteLayout").getDeviceType() === "smartphone") {
				regularPanel = "header";
				panelInfo = {
					name: "headerMultisel",
					topIndex: 0,
					selector: "header",
					$type: "application/x-panel-header-multisel-smartphone"
				}
			} else {
				regularPanel = "footer";
				panelInfo = {
					name: "footerMultisel",
					topIndex: 0,
					selector: "footer",
					$type: "application/x-panel-footer-multisel-tablet"
				}
			}
			this._multiSelArray = this.controller.getControlsByType("application/x-array");
			if (this._multiSelArray.length !== 1) {
				throw new Error("One and only one array is expected in a query facet")
			}
			this._multiSelArray = this._multiSelArray[0];
			this._multiselPanel = this.attachedControlsMgr.findByName(panelInfo.name);
			if (!this._multiselPanel) {
				this._multiselPanel = this.attachedControlsMgr.createAddtionalControl(panelInfo, {});
			}
			this.mutiSelectionEnabled = this.attachedControlsMgr.toggleControls(panelInfo.name, regularPanel);
			this._multiselPanel.multiSelToggleStatus(this.mutiSelectionEnabled, this._multiSelArray);
			if (!this.mutiSelectionEnabled) {
				this.$$contentElmt.find(".s-m-multi-selected").removeClass("s-m-multi-selected");
			}
		},
		multiSelToggleRecord: function($$elmt) {
			this._multiselPanel.multiSelToggleRecord($$elmt);
		},
		_actMultiSelDownload: function() {
			var self = this;
			alert("_startDownloadAction");
		}
	});
