// syracuse-tablet/html/js/wc/pages/sdata/pageHome.js
"use strict";

var Base = require('syracuse-tablet/html/js/wc/pages/sdata/pageSdataDashboard').Page;
var BaseSdata = require('syracuse-tablet/html/js/wc/pages/sdata/pageSdata').Page;
var utils = require('syracuse-tablet/html/js/wc/helpers/utils');
var locale = require('syracuse-tablet/html/js/wc/helpers/locale');
var HomeController = require("syracuse-tablet/html/js/wc/controllers/homeController").HomeController;
var native = require('syracuse-tablet/html/js/wc/helpers/native/native');
var wpHelpers = require('syracuse-tablet/html/js/wc/sdata/wpHelpers');
var modules = require('syracuse-tablet/html/js/wc/common/modules');

/**
 *
 */
exports.Page = utils.defineClass(
	function PageHome(pageData, options) {
		Base.call(this, pageData, options);
		this.actionAdapter.showDesignPage = false;
		this.actionAdapter.showPageShare = false;
		this.actionAdapter.showMailTo = false;
		this.actionAdapter.showPinPage = false;
		this.actionAdapter.showHome = false;
		this.actionAdapter.showPageConfig = true;
		this.nativeVoiceCommands = native.getModule("voiceCommands");
	},
	Base, {
		destroy: function() {
			Base.prototype.destroy.call(this);
		},
		getPageTitle: function() {
			return locale.text("global.pageTitle.home");
		},
		isRootPage: function() {
			return true;
		},
		isPersonal: function() {
			return this.pageData.isPersonal;
		},
		_ensureController: function() {
			this.controller = new HomeController(this.pageData.dataset, this);
			// Will raise notifications or do confirmations
			this.controller.setUIAdapter(this);
			this.controller.setNativeCapabilities(this.options)
		},

		afterRender: function() {
			Base.prototype.afterRender.call(this);
			this.updateVoiceCommands();

		},

		updateVoiceCommands: function() {
			var self = this;
			if (self.nativeVoiceCommands) {
				self.getVoiceCommands().then(function(commands) {
					self.nativeVoiceCommands.updateVoiceCommands(commands);
				});
			}
		},
		_actPageConfig: function() {
			this.controller.toggleConfig(true);
		},
		_enabledConfig: function(forceShow) {
			this.controller.enabledConfig(forceShow)
			var $$tiles = $(".s-m-tile", this.$$elmt);
			$$tiles.addClass("s-m-tile-edit")
		},
		_disabledConfig: function() {
			this.controller.disabledConfig();
			$(".s-m-tile-edit", this.$$elmt).removeClass("s-m-tile-edit");
		},
		/**
		 * List of voice commands that can be called by user
		 * will be stored in native api and open the called page if the title matches
		 * 
		 * Promise resolving with:
		 * [
		 * {
		 *   title: "My Customers",
		 *   page: {
		 *     $url: ...
		 *     $method: ...
		 *   }
		 * ]
		 */
		getVoiceCommands: function() {
			return $.smResolve(wpHelpers.welcomePageToCommands(this.pageData));
		},
		getPageDescription: function() {
			return this.pageData.meta.description
		},
		_actCloseDashboardConfig: function() {
			this.controller.toggleConfig();
			this._refresh(false)
		},
		_actSavePersonal: function() {
			var self = this;
			var data = wpHelpers.createWelcomeDashboardFromPageData(this.pageData, true);
			wpHelpers.saveWelcomeDashboard(data).then(function(res) {
				self.controller.cleanContext();
				modules.get("modal").notify({
					body: locale.text("dashboard.welcome.saved"),
					severityClass: "success"
				});
			})
		},
		_actAddGroup: function() {
			// Called when add group in the footer is clicked
			this.pageData.page.$article.$items.push({
				$items: [],
				$layoutType: "hubgroup",
				$title: locale.text("welcome.newgroup")
			})
			this._refresh(true)
		},
		_refresh: function(dirty) {
			var self = this;
			self.authUpdateLayout(self.pageData.page.$article)
				.then(function() {
					if (self.controller.configActive) { // config already enabled restore it
						self._enabledConfig(true)
					}
					self.controller.setDirty(dirty)
				})
				.fail(function(e) {
					modules.get("modal").error(e);
				});
		},
		authUpdateLayout: function(article) {
			var self = this;
			return BaseSdata.prototype.authUpdateLayout.call(self, article)
		},
	}
);
