// @sage/syracuse-ui/lib/links/menusHelper.js
"use strict";
exports.click = require('./clickHandlers');
var _quickEditBuilder = require("syracuse-ui/lib/field/array/gridMode/tools/quickEditBuilder");
var _paging = require("syracuse-ui/lib/field/array/tools/paging");
var _lookup = require("syracuse-ui/lib/field/reference/lookup");
var _tunnel = require("syracuse-ui/lib/field/reference/tunnel");

function _clear(item) {
	if (item.buttons) {
		for (var ii = 0, jj = item.buttons.length; ii < jj; ii++) {
			var btn = item.buttons[ii];
			btn.pickerId && syra_button.remove(btn);
		}
	}
	var $binds = Object.keys(item.menuItems);
	for (var ii = 0, jj = $binds.length; ii < jj; ii++) {
		var $bind = $binds[ii];
		var mn, items = item.menuItems[$bind];
		for (var kk = 0, mm = items.length; kk < mm; kk++) {
			mn = items[kk];
			if (mn.detailitemId) {
				var item = syra_item.get(mn.detailitemId);
				item && item.setValue(item.currentValue);
			}
			if (mn.layoutParent) {
				mn.layoutParent.removeItem(mn, true);
			} else {
				syra_item.remove(mn, true);
			}
		}
	}
	exports.initializeScope(item);
	if (item.menusPopupBox) {
		if (item.menusPopupBox.layoutParent) {
			item.menusPopupBox.layoutParent.removeItem(item.menusPopupBox, true);
		} else {
			syra_item.remove(item.menusPopupBox, true);
		}
		if (item.menusBag == item.menusPopupBox) {
			delete item.menusBag;
		}
		delete item.menusPopupBox;
	}
	syra_button.hide(item.picker_menus, true);
}

function _applyMenusChange(field, $deltaMenu, $isAction, metaData) {
	var binds = Object.keys($deltaMenu);
	for (var ii = 0, jj = binds.length; ii < jj; ii++) {
		var $bind = binds[ii];
		var $menu = $deltaMenu[$bind];
		var created;
		if ($menu === null) {
			if (field.menuItems) {
				var menuItems = field.menuItems[$bind];
				if (menuItems) {
					for (var mm = 0, kk = menuItems.length; mm < kk; mm++) {
						syra_item.remove(menuItems[mm]);
					}
					delete field.menuItems[$bind];
				}
			}
			delete field.$menus[$bind];
			delete field.$menusPopup[$bind];
		} else {
			if ($menu.$target == "embedded") {
				syra_fields.embedded.addEmbeddedLink(field, $bind, $menu);
				continue;
			}
			var $defined = field.$menus[$bind];
			if ($defined) {
				syra_dataset.applyDelta(field.page, $defined, $menu);
			} else {
				created = true;
				$menu.$isAction = $isAction;
				$menu.$sourceBind = $menu.$sourceBind || $bind;
				$menu.$bind = field.isField ? (field.id + "-" + $bind) : $bind;

				//$defined = field.$menus[$bind] = $menu;
				$defined = field.$menus[$bind] = {};
				syra_dataset.applyDelta(field.page, $defined, $menu);
				switch ($bind) {
					case "$first":
					case "$previous":
					case "$next":
					case "$last":
						field.isDesktopPage && _paging.addPagerLinksToPage(field, $bind);
						break;
					case "$edit":
						if (field.onAddMenuItem) {
							if (field.articleParent.$capability && field.articleParent.$capability.quickedit) {
								_quickEditBuilder.addPicker(field, $defined);
								$defined.$title = syra_local.fieldTunnel;
								$defined.$iconValue = "$tunnel";
								$defined.$isPopup = true;
							} else {
								$defined.$title = $defined.$title || syra_local.fieldEdit;
								field.onAddMenuItem($bind, $defined);
								$defined.$isPopup = false;
							}
						} else {
							$defined.$isPopup = true;
						}
						break;
					case "$canceledit":
						if (field.onAddMenuItem) {
							$defined.$title = $defined.$title || syra_local.fieldEdit;
							field.onAddMenuItem($bind, $defined);
							$defined.$isPopup = false;
						} else {
							$defined.$isPopup = true;
						}
						break;
					case "$delete":
						if (field.onAddMenuItem && field.$useDeletePicker) {
							$defined.$title = $defined.$title || syra_local.fieldDelete;
							field.onAddMenuItem($bind, $defined);
							$defined.$isPopup = false;
						} else {
							$defined.$isPopup = true;
						}
						break;
					case "remove_separator":
					case "add_separator":
						field.onAddMenuItem && field.onAddMenuItem($bind, $defined);
						$defined.$isPopup = false;
						break;
					case "$refresh":
						$defined.$isPopup = !field.page.isVignettePage;
						break;
					case "$create":
					case "$select":
						$defined.$isPopup = !(field.isArrayField && field.onAddMenuItem($bind, $defined));
						break;
					case "$details":
						$defined.$isPopup = !field.isField || field.$isDetailLinkDisabled;
						break;
					case "$tunnel":
						_tunnel.addPicker(field, $defined);
						break;
					case "$lookup":
						_lookup.addPicker(field, $defined);
						break;
					case "$localize":
						syra_localizer.addPicker(field, $defined);
						break;
					case "menus":
						exports.picker.add(field);
						break;
					default:
						$defined.$isPopup = true;
						break;
				}

			}
			if ($defined.$isPopup && field.$menusPopup) {
				if ($menu.$isHidden) {
					delete field.$menusPopup[$bind];
				} else {
					field.$menusPopup[$bind] = $defined;
				}
			}
			if (field.menuItems) {
				var bounds = field.menuItems[$bind];
				if (bounds) {
					//if create setMenu is already done
					for (var mm = 0, kk = bounds.length; mm < kk; mm++) {
						bounds[mm].setMenu($menu, field.dataset);
					}
				} else {
					field.menusBag && field.menusBag.addMenuItem($menu);
				}
			}
		}
	}
}


exports.initializeScope = function(scope) {
	scope.$menus = {};
	scope.$menusPopup = {};
	scope.menuItems = {};
};

exports.applyChange = function(field, metaData) {
	if (metaData) {
		if (field.variantItem) {
			field = field.variantItem;
		}
		if (field.isChildField) {
			exports.applyChange(field.record, {
				$links: metaData.$links,
				$actions: metaData.$actions
			});
			return;
		}
		if (field.renderIcon) {
			metaData.$links !== undefined && field.renderIcon(metaData);
			return;
		}
		// in order not to show menu menu picker for unit ref field        
		if (field.$item && !field.$item.$isMenusHidden && (!field.$item.$isUnitField || field.$isEditMode)) {
			var $menus = (!field.isArrayField && metaData.$item) || metaData;
			if ($menus.$links !== undefined || $menus.$actions !== undefined) {
				if ($menus.$links === null) {
					_clear(field);
				} else {
					$menus.$links && _applyMenusChange(field, $menus.$links, false, metaData);
					$menus.$actions && _applyMenusChange(field, $menus.$actions, true, metaData);
					if (!field.disposed && field.$item) {
						//!!check disposed as field can be disposed during apply ( field is page and menu is save, closing dlg)
						if (!field.picker_menus && !field.$item.$isFilterMode && !field.$item.$isPickerMenuHidden && field.$menusPopup && Object.keys(field.$menusPopup).length) {
							exports.picker.add(field);
							if (!field.$isDisabled && field.$isReadOnly) {
								syra_button.hide(field.picker_menus, false);
							}
						}
					}
				}
			}
		}
	}
};


function _disableItem(menuItem, isDisable) {
	menuItem && menuItem.disable(isDisable);
}

exports.disableItems = function(article, menuItems) {
	if (menuItems) {
		var isDisable = article.isPickersDisabled = article.$isReadOnly || article.$isDisabled;
		menuItems.$quickEdit && _disableItem(menuItems.$quickEdit[0], isDisable);
		menuItems.$edit && _disableItem(menuItems.$edit[0], isDisable);
		menuItems.$delete && _disableItem(menuItems.$delete[0], isDisable);
		menuItems.$create && _disableItem(menuItems.$create[0], isDisable);
		menuItems.$select && _disableItem(menuItems.$select[0], isDisable);
	}
};

exports.applyFieldState = function(field) {
	if (field.buttons) {
		for (var ii = 0, jj = field.buttons.length; ii < jj; ii++) {
			var btn = field.buttons[ii];
			if (btn.pickerId) {
				if (btn.pickerId == "$tunnel" || btn.pickerId == "menus") {
					syra_button.hide(btn, field.$isDisabled);
				} else {
					syra_button.hide(btn, field.$isDisabled || field.$isReadOnly);
				}
			}
		}
	}
	exports.disableItems(field, field.menuItems);
};

exports.disposeScope = function(scope) {
	scope.menusPopup && scope.menusPopup.close();
	exports.click.dispose(scope);
};


var _listMenuBinds = ["$edit", "$create", "$delete", "$details", "$design"];
var _popupArticle;

exports.open = function(options) {
	if (_popupArticle) {
		syra_item.remove(_popupArticle, true);
		_popupArticle = null;
	}
	var scope = options.scope;
	if (!scope && options.$scope) {
		_popupArticle = scope = syra_item.addRecord(options.$scope);
	}
	if (scope) {
		if (!scope.menusPopup) {
			var $item = {
				$noText: true,
				$title: syra_local.fieldActions,
				$skin: "s-mn-popup",
				$itemSkin: options.$itemSkin || "s-list-default-btn-default",
				$category: "menus",
				$layout: {
					$items: []
				}
			};
			var article = (scope.isDesktopPage || scope.isRecordArticle) ? scope : scope.articleParent;
			if (scope.menusPopupBox && article.menuItems) {
				//clear previous popup menus 
				var items = scope.menusPopupBox.boxChildItems;
				for (var ii = 0, jj = items.length; ii < jj; ii++) {
					delete article.menuItems[items[ii].$bind];
					syra_item.remove(items[ii]);
				}
				syra_item.remove(scope.menusPopupBox, true, true);
			}
			scope.menusPopupBox = scope.page.addItem(null, $item, article);
			scope.menusPopupBox.contextField = scope;
			if (options.onMenuClick) {
				scope.menusPopupBox.onMenuClick = options.onMenuClick;
			}
			var binds = Object.keys(scope.$menusPopup);
			//setOrder
			var newBinds = [];
			for (var ii = 0, jj = _listMenuBinds.length; ii < jj; ii++) {
				var found = binds.indexOf(_listMenuBinds[ii]);
				if (found >= 0) {
					binds.splice(found, 1);
					newBinds.push(_listMenuBinds[ii]);
				}
			}
			binds = newBinds.concat(binds);
			for (var ii = 0, jj = binds.length; ii < jj; ii++) {
				var $menu = scope.$menusPopup[binds[ii]];
				if (!$menu.$icon) {
					if (_listMenuBinds.indexOf($menu.$sourceBind) >= 0) {
						$menu.$icon = {
							$mode: "iconText",
							$path: scope.$iconPath
						};
					}
				}
				var menuItem = scope.menusPopupBox.addMenuItem($menu);
				article.$menus[$menu.$bind] = $menu;
				if (!menuItem.cssTitle) {
					//ensure padding for no icon            
					menuItem.domItem.className += (menuItem.cssTitle = " s-popup-menu-no-icon");
				}
				menuItem.setMenu($menu, article.dataset);
			}
			syra_menus.disableItems(scope, scope.menuItems);
			if (!options.picker) {
				return;
			}
			var picker = options.picker;
			if (picker.pickerId) {
				syra_button.setText(picker, undefined, "menus_open");
				picker = picker.link;
			} else {
				if (picker.link) {
					picker = picker.link;
				}
			}
			syra_quality.openPopupMenu(scope.menusPopupBox);
			scope.menusPopup = syra_popup.open({
				slot: scope.menusPopupBox.body,
				my: "right top",
				at: "right bottom",
				of: picker,
				close: function() {
					if (picker.pickerId) {
						syra_button.setText(picker, undefined, "menus");
					}
					scope.menusPopup = null;
				}
			});
		} else {
			scope.menusPopup.close();
		}
	}
};

exports.picker = {
	add: function(field) {
		return syra_button.addFieldPicker({
			parent: field,
			pickerId: "menus",
			fontIcon: "menus",
			text: syra_local.flActions,
			shortCutTip: (field.isField && !field.isArrayField) ? syra_shortcut.tip.contextMenu : null,
			click: this.onClick
		});
	},
	onClick: function() {
		syra_menus.open({
			scope: this.parent,
			picker: this
		});
	},
	show: function(field) {
		if (field.picker_menus) {
			var show = field.$isEditMode || field.hasValue();
			syra_button.hide(field.picker_menus, !show);
			if (!field.picker_menus.link.parentNode && field.onAddMenuPicker) {
				field.onAddMenuPicker(field.picker_menus);
			}
		}
	}
};
