// syracuse-tablet/html/js/controls/panels/panelBase.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var Base = require('syracuse-tablet/html/js/controls/ctrlBase').CtrlBase;

/**
 * 
 */
exports.PanelBase = utils.defineClass(function PanelBase(controller, type, article, opts) {
	Base.call(this, controller, type, article, opts);
	this.isPanel = true;
}, Base, {
	destroy: function() {
		Base.prototype.destroy.call(this);
		// Do not call destroy since we did not create the adapter but just reference to it
		this.actionAdapter = null;
	},

	/**
	 * This must be called by the creator if this component (usually it's the page)
	 * The component will use the adapter instance to get a list of items to be rendered
	 */
	setActionAdapter: function(actionAdapter) {
		this.actionAdapter = actionAdapter;
	},

	//
	// START: Rendering
	// 
	buildHtml: function() {
		this.createRootElement(["s-m-panel", "s-m-control", "s-m-ctrl"]);
	},
	//
	// END: Rendering
	//

	createId: function(typeName) {
		return utils.readableuid(typeName);
	},
	/**
	 * Built with the rank of the layout in generation process to be reliable
	 */
	getUnitTestId: function() {
		if (this.options.index == null) return;
		return utils.unitTestId(this.typeName.toLowerCase(), this.options.index);
	},
	getGestureMgr: function() {
		return null;
	},
	/**
	 * disable/enable clicks on the panel
	 */
	disable: function(yes) {
		yes = yes === true;
		this.$$elmt.css("opacity", yes ? 0.5 : "inherit");
		if (yes) {
			this.$$elmt.click(function(evt) {
				evt.preventDefault();
				evt.stopPropagation();
			})
		} else {
			this.$$elmt.off();
		}
	},
	_action$$Elmt: function(action, params) {
		var selector = '[data-naction="' + action + '"]';
		if (params) {
			selector += '[data-params=' + params + ']';
		}
		var res = $();
		this.$$elmt.find(selector).each(function() {
			var $$elmt = $(this);
			res.push($$elmt);
			$$elmt.closest('[data-menuitem-id="' + action + '"]').each(function() {
				res.push($(this))
			})
		})
		return res;

	},
	enableAction: function(action, yes, params) {
		this._action$$Elmt(action, params).each(function() {
			$(this).toggleClass("disabled", yes === false);
			if (yes) {
				$(this).removeAttr("disabled");
			} else {
				$(this).attr("disabled", "true");
			}
		})
	},
	showAction: function(action, yes, params) {
		this._action$$Elmt(action, params).each(function() {
			$(this).toggle(yes !== false);
		})
	}
});
