// msoffice/lib/ppt/pptInterfaceDummy.js
"use strict";

// Used by pptInterface if Ppt is not opened (testing in browser)
// Test URLS:
// http://localhost:8124/msoffice/lib/ppt/ui/main.html?url=http%3A%2F%2Flocalhost%3A8124%2Fsdata%2Fsyracuse%2Fcollaboration%2Fsyracuse%2Fusers%3Frepresentation%3Duser.%24query%26role%3D98184048-6790-480f-b1c5-73e6268e51c8
var helpers = require('syracuse-core/lib/helpers');
var msofficeHelpers = require('msoffice/lib/helpers');

function PptInterfaceDummy() {
	var self = this;
}
exports.PptInterfaceDummy = helpers.defineClass(PptInterfaceDummy, null, {

	getResourceUrl: function() {
		var val = this._getBaseURIParam("url");
		return val;
	},
	addDataToWorksheet: function(data) {
		var jsonData = JSON.stringify(data);
		alert(jsonData);
	},
	getDocumentContent: function() {
		return "Dummy content";
	},
	getDocumentTitle: function() {},
	setDocumentTitle: function(value) {},
	getDocumentUrl: function() {},
	setDocumentUrl: function(value) {},
	refreshNextChart: function() {},
	notifySaveDocumentDone: function(value) {},
	getActionType: function() {
		return "ppt_populate_worksheet";
	},
	getAddinVersion: function() {},
	_getBaseURIParam: function(param) {
		var params = document.baseURI.split("?")[1].split("&");
		var val;
		params.forEach(function(p) {
			var idx = p.indexOf(param + "=");
			if (idx > -1) {
				val = p.substring(idx + param.length + 1);
				val = decodeURIComponent(val);
			}
		});
		return val;
	},
	expectedVersion: function(value) {}
});
