// msoffice/lib/ppt/pptInterfaceImpl.js
"use strict"

// Used by pptInterface if Ppt is opened

var helpers = require('syracuse-core/lib/helpers');
var msofficeHelpers = require('msoffice/lib/helpers');
var base64 = require("msoffice/lib/Base64").Base64;

function PptInterfaceImpl() {
}

exports.PptInterfaceImpl = helpers.defineClass(PptInterfaceImpl, null, {
	getResourceUrl: function () {
		return external.getPptCustomXlsData().getResourceUrl();
	},
	addDataToWorksheet: function(data) {
		var jsonData = JSON.stringify(data);
		external.addDataToWorksheet(jsonData);
	},
	addDataToWorksheetFinished: function() {
		external.addDataToWorksheetFinished();
	},
	getDocumentContent: function() {
        return base64.raw_decode(external.GetDocumentContent());
	},
	getDocumentTitle: function() {
		return external.getPptCustomData().getDocumentTitle();
	},
	setDocumentTitle: function(value) {
		external.getPptCustomData().setDocumentTitle(value);
		external.getPptCustomData().writeDictionaryToDocument();
	},
	getDocumentUrl: function() {
		return external.getPptCustomData().getDocumentUrl();
	},
	setDocumentUrl: function(value) {
		external.getPptCustomData().setDocumentUrl(value);
		external.getPptCustomData().writeDictionaryToDocument();
	},
	notifySaveDocumentDone: function(value) {
		external.NotifySaveDocumentDone();	
	},
	getActionType: function() {
		var actionType = external.getPptCustomData().getActionType();
		return actionType;
	},
	getAddinVersion: function() {
		return external.getAddinVersion();
	}
});
