// msoffice/lib/ppt/pptInterfaceDummy.js
"use strict"

// Used by pptInterface if Ppt is not opened (testing in browser)

// Test URLS:
// http://localhost:8124/msoffice/lib/ppt/ui/main.html?url=http%3A%2F%2Flocalhost%3A8124%2Fsdata%2Fsyracuse%2Fcollaboration%2Fsyracuse%2Fusers%3Frepresentation%3Duser.%24query%26role%3D98184048-6790-480f-b1c5-73e6268e51c8
var helpers = require('syracuse-core/lib/helpers');
var msofficeHelpers = require('msoffice/lib/helpers');

function PptInterfaceDummy() {
	var self = this;
}
exports.PptInterfaceDummy = helpers.defineClass(PptInterfaceDummy, null, {
	
	getResourceUrl: function () {
		//Test URLS
		//  list of users
		return "/sdata/syracuse/collaboration/syracuse/users?representation=users.$query";
	},
	addDataToWorksheet: function(data) {
		var jsonData = JSON.stringify(data);
		alert(jsonData);
	},
	getDocumentContent: function() {
        return "Dummy content";
	},
	getDocumentTitle: function() {
	},
	setDocumentTitle: function(value) {
	},
	getDocumentUrl: function() {
	},
	setDocumentUrl: function(value) {
	},
	notifySaveDocumentDone: function(value) {
	},
	getActionType: function() {
		return "ppt_populate_worksheet";
	},
	getAddinVersion: function() {
	}
});
