// syracuse-ui/lib/site/authoring/palette/fieldPalette.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var ItemPalette = require("./itemPalette").ItemPalette;

function FieldPalette(){
}

exports.FieldPalette = helpers.defineClass(FieldPalette, ItemPalette, {
    loadBox: function(){
        var localize = this.authorPage.localize;
        this.$prototype = {
            "$properties": {
                $isHidden: {
                    $type: "application/x-boolean",
                    $title: localize.aw_isHidden
                },
                $isTitleHidden: {
                    $type: "application/x-boolean",
                    $title: localize.aw_isLabelHidden
                }
            },
            $isTitleHidden: false
        };
        this._formatFieldLabelAlign(true);
        this._formatFieldTextLabelAlign(true);
        this.authorPage.initializeNewItem(this, {
            $category: "section",
            $bind: "$field",
            $skin: "s-aw-designer-item",
            $layout: {
                $layoutType: "row",
                $fitToContent: true,
                $items: [{
                    $category: "section",
                    $title: localize.aw_appearance,
                    $layout: {
                        $layoutType: "row",
                        $fitToContent: true,
                        $items: [{
                            $layoutType: "stack",
                            $items: [{
                                $bind: "$isHidden",
                                $skin: "s-aw-field",
                                $isEditMode: true,
                                $isLeftSpaceHidden: true
                            }, {
                                $bind: "$isTitleHidden",
                                $skin: "s-aw-field",
                                $isEditMode: true,
                                $isLeftSpaceHidden: true
                            }]
                        }, {
                            $category: "block",
                            $title: localize.aw_fieldLabelPosition,
                            $layout: {
                                $items: [this._formatFieldLabelAlign(false), this._formatFieldTextLabelAlign(false)]
                            }
                        }]
                    }
                }, {
                    $category: "block",
                    $layout: {
                        $items: []
                    }
                }]
            
            }
        }, this.authorPage);
        ItemPalette.prototype.loadBox.call(this);
        if (this.getNode().item.getAuthoringWidget) {
            if (this.fieldAuthoring = this.getNode().item.getAuthoringWidget()) {
                this.fieldAuthoring.$$container = this.layoutContent.items[1].$$body;
                document.itemFactory.initializeItem(this.fieldAuthoring, {}, this);
                this.fieldAuthoring.designedField = this.getNode().item;
                this.fieldAuthoring.$layoutOptions = document.site.$item.$layoutOptions.authoring;
                this.fieldAuthoring.$skin = this.fieldAuthoring.$layoutOptions.section;
                this.fieldAuthoring.loadBox(this.getNode().item.$item);
            }
        }
    },
    dispose: function(){
        if (this.fieldAuthoring) {
            delete this.fieldAuthoring.designedField;
            delete this.fieldAuthoring;
        }
        ItemPalette.prototype.dispose.call(this);
    }
});
