// syracuse-ui/lib/site/authoring/palette/itemPalette.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var RawPage = require("syracuse-ui/lib/article/rawPage").RawPage;
var layoutSettings = require('syracuse-ui/lib/article/layouts/authoring/layoutSettings');

function ItemPalette(){
}

exports.ItemPalette = helpers.defineClass(ItemPalette, RawPage, {
    getNode: function(){
        return this.authorPage.designedNode;
    },
    isFieldItem: function(){
        return this.getNode().$authoringLevel == "field";
    },
    _notifyChildrenFields: function(children, value, binding){
        var self = this;
        children.forEach(function(child){
            if (child.$authoringLevel == "field") {
                self._notifyFieldItem(child.item, value, binding);
            }
            else {
                if (child.children) {
                    self._notifyChildrenFields(child.children, value, binding);
                }
            }
        });
    },
    _notifyFieldItem: function(item, value, binding){
        var metaData = {}
        metaData[binding] = value;
        item.applyDesignMetaData(metaData, true);
        if (item.authoringNode && metaData.$isHidden !== undefined) {
            item.authoringNode.applyHiddenState(metaData);
        }
    },
    notifyDataChange: function(field, value){
        var self = this;
        var node = self.getNode();
        var fieldArticle = field.getArticle();
        if (fieldArticle.designedField) {
            var metaData = {};
            metaData[field.$item.$bind] = value;
            fieldArticle.designedField.applyDesignMetaData(metaData, true);
            fieldArticle.applyChange(metaData);
            if (fieldArticle.onNotifyAuthoringChange) {
                fieldArticle.onNotifyAuthoringChange(field, value, metaData);
            }
            self.authorPage.notifyUpdate();
        }
        else {
            var metaData = {};
            metaData[field.$item.$bind] = value;
            if (self.isFieldItem()) {
                self._notifyFieldItem(node.item, value, field.$item.$bind);
            }
            else {
                if (field.$item.$bind.indexOf("$fields") == 0) {
                    var fieldBind = field.$item.$bind.slice("$fields".length);
                    fieldBind = fieldBind.slice(0, 1).toLowerCase() + fieldBind.slice(1);
                    self._notifyChildrenFields(node.children, value, "$" + fieldBind);
                    node.item.applyDesignMetaData(metaData);
                }
                else {
                    if (metaData.$boxLayoutType !== undefined) {
                        metaData.$layout = {};
                        if (metaData.$boxLayoutType == "100") {
                            metaData.$layout.$layoutType = self.boundFields.$allChildLayout[0].currentValue;
                        }
                        else {
                            metaData.$layout.$layoutType = "columns";
                            metaData.$layout.$layoutSubType = metaData.$boxLayoutType.replace(/-/g, ",");
                        }
                    }
                    if (metaData.$allChildLayout) {
                        if (metaData.$allChildLayout == "tabs") {
                            if (this.getNode().getChildrenLevel() == "field") {
                                var authorPage = this.authorPage;
                                var newBox = authorPage.addNewBox({
                                    parentBox: node.item,
                                    $category: node.$authoringLevel == "section" ? "block" : "section",
                                    doSelectItem: false
                                });
                                authorPage.moveContent({
                                    sourceNode: node,
                                    targetItem: newBox
                                });
                                metaData.$layout = metaData.$layout || {};
                                metaData.$layout.$layoutType = metaData.$allChildLayout;
                                delete metaData.$allChildLayout;
                                node.item.applyDesignMetaData(metaData);
                                return null;
                            }
                        }
                        metaData.$layout = metaData.$layout || {};
                        metaData.$layout.$layoutType = metaData.$allChildLayout;
                        delete metaData.$allChildLayout;
                    }
                    node.item.applyDesignMetaData(metaData);
                    if (metaData.$isHidden !== undefined) {
                        node.applyHiddenState(metaData);
                    }
                }
            }
            self.authorPage.notifyUpdate();
            self.applyChange(metaData);
        }
    },
    applyChange: function(newData){
        var self = this;
        var metaData = {
            $properties: {}
        };
        if (newData) {
            Object.keys(newData).forEach(function($prop){
                var value = newData[$prop];
                if ($prop == "$layout") {
                    $prop = "$boxLayoutType";
                    switch (newData.$layout.$layoutType || "stack") {
                        case "row":
                        case "stack":
                        case "side":
                        case "tabs":
                            value = "100";
                            metaData.$allChildLayout = newData.$layout.$layoutType;
                            break;
                        default:
                            value = (newData.$layout.$layoutSubType || "50,50").replace(/,/g, "-");
                            break;
                    }
                }
                switch ($prop) {
                    case "$boxLayoutType":
                        if (!newData.$allChildLayout && !metaData.$allChildLayout) {
                            metaData.$allChildLayout = "stack";
                        }
                        metaData.$properties.$allChildLayout = {
                            $isDisabled: value != "100"
                        };
                        break;
                    case "$isTopLabelAlignment":
                        metaData.$properties.$isRightTextLabelAlignment = {
                            $isDisabled: value
                        };
                        break;
                    case "$fieldsIsTopLabelAlignment":
                        metaData.$properties.$fieldsIsRightTextLabelAlignment = {
                            $isDisabled: value
                        };
                        break;
                    case "$isHidden":
                        var $isDisabled = {
                            $isDisabled: value
                        };
                        Object.keys(self.$prototype.$properties).forEach(function($key){
                            if ($key.indexOf("$isHidden") < 0) {
                                metaData.$properties[$key] = $isDisabled;
                            }
                        });
                        if (self.menuItems) {
                            ["$switchMode", "$addBlock", "$addSection", "$addField"].forEach(function(prop){
                                var menus = self.menuItems[prop];
                                if (menus) {
                                    menus.forEach(function(menu){
                                        menu.setMenu($isDisabled);
                                    });
                                }
                                
                            });
                        }
                        break;
                    case "$fieldsIsTitleHidden":
                        (metaData.$properties.$fieldsIsTopLabelAlignment = metaData.$properties.$fieldsIsTopLabelAlignment || {}).$isDisabled = value;
                        var isRightDisabled = !value ? (newData.$fieldsIsTopLabelAlignment || self.dataset.$fieldsIsTopLabelAlignment) : true;
                        (metaData.$properties.$fieldsIsRightTextLabelAlignment = metaData.$properties.$fieldsIsRightTextLabelAlignment || {}).$isDisabled = isRightDisabled;
                        break;
                    case "$isTitleHidden":
                        if (self.isFieldItem()) {
                            (metaData.$properties.$isTopLabelAlignment = metaData.$properties.$isTopLabelAlignment || {}).$isDisabled = value;
                            var isRightDisabled = !value ? (newData.$isTopLabelAlignment || self.dataset.$isTopLabelAlignment) : true;
                            (metaData.$properties.$isRightTextLabelAlignment = metaData.$properties.$isRightTextLabelAlignment || {}).$isDisabled = isRightDisabled;
                        }
                        else {
                            metaData.$properties.$title = metaData.$properties.$isBoxCollapsable = metaData.$properties.$isMaximizable = {
                                $isDisabled: value
                            };
                        }
                        break;
                }
                metaData[$prop] = value;
            });
        }
        RawPage.prototype.applyChange.call(this, metaData);
    },
    _setSwitchMode: function(){
        var self = this;
        this.menuItems.$switchMode.forEach(function(menu){
            menu.setTitle(self.authorPage.$isLightMode ? self.localize.aw_fullMode : self.localize.aw_showLightMode);
            menu.setValue(self.authorPage.$isLightMode ? "switch-full" : "switch-light");
        });
    },
    onAddFields: function(menuItem){
        var self = this;
        if (!self._popupAdFields) {
            var targetItem, action;
            var div = document.createElement("div");
            div.className = "s-aw-fields-pn";
            self.$$popupPanel = $(div).delegate("a.s-aw-fields-field-value", "click", function(){
                var $bind = this.getAttribute("data-s-field");
                setTimeout(function(){
                    var node = self.getNode();
                    var layoutParent;
                    if (!self.isFieldItem()) {
                        var $childLevel = node.getChildrenLevel();
                        if ($childLevel == "section" || $childLevel == "block") {
                            layoutParent = self.authorPage.addNewBox({
                                parentBox: node.item,
                                $category: $childLevel || "section",
                                doSelectItem: false,
                                openBox: true
                            }).layoutContent;
                        }
                        else {
                            layoutParent = node.item.layoutContent;
                        }
                    }
                    else {
                        layoutParent = node.item.layoutParent;
                        targetItem = node.item;
                        action = "insertAfter";
                    }
                    layoutParent.appendNewItem({
                        load: true,
                        newItem: layoutParent.box.page.createNewItem({
                            $bind: $bind
                        }, layoutParent.box),
                        targetItem: targetItem,
                        action: action
                    });
                    if (self._popupAdFields) {
                        self._popupAdFields.close();
                    }
                    document.site.authorPage.notifyUpdate();
                    
                }, 10);
                return false;
            });
            var domTable = document.createElement("div");
            domTable.className = "s-aw-fields-table";
            self.$$popupPanel.append(domTable);
            
            var article = self.getNode().item.getArticle();
            if (!article.$prototype.$properties) {
                article = article.getArticleParent();
            }
            var domCol, colLength;
            var excludeBinds = ["$creUser", "$updUser", "$creDate", "$updDate"];
            var $properties = article.$prototype.$properties;
            Object.keys($properties).forEach(function($name){
                if (excludeBinds.indexOf($name) < 0) {
                    var $field = $properties[$name];
                    if ($field.$type) {
                        if (!domCol) {
                            colLength = 0;
                            domCol = document.createElement("div");
                            domCol.className = "s-aw-fields-col";
                            domTable.appendChild(domCol);
                        }
                        var domValue = document.createElement("a");
                        domValue.className = "s-aw-fields-field-value";
                        domValue.setAttribute("data-s-field", $name);
                        
                        var div = document.createElement("div");
                        div.className = "s-aw-fields-field-type";
                        div.style.backgroundImage = "url('" + document.site.$item.$iconPath + "authoring/s-aw-light-" + $field.$type.replace("application/x-", "").replace("/", "-") + ".png')";
                        domValue.appendChild(div);
                        
                        div = document.createElement("div");
                        div.className = "s-aw-fields-field-title";
                        $(domValue.appendChild(div)).text($field.$title ? article.getLocalizeText($field.$title) : self.localize.aw_defaultFieldTitle);
                        
                        div = document.createElement("a");
                        div.className = "s-aw-fields-field-item";
                        div.appendChild(domValue);
                        domCol.appendChild(div);
                        if (++colLength == 10) {
                            domCol = null;
                        }
                    }
                    else {
                        console.log("field type not defined:" + $name);
                    }
                }
            });
            
            menuItem.$$item[0].setAttribute("id", menuItem.id);
            self._popupAdFields = self.openDialog({
                $dialogMode: "popup",
                content: self,
                $$dialog: self.$$popupPanel,
                position: {
                    my: "right top",
                    at: "right bottom",
                    of: menuItem.$$item
                },
                onClose: function(){
                    self.$$popupPanel.undelegate();
                    self._popupAdFields = null;
                },
                autocCloseBoundary: "#" + menuItem.id
            });
        }
        else {
            self._popupAdFields.close();
        }
    },
    onMenuClick: function(menuItem){
        if (menuItem.$bind.indexOf("composite") >= 0) {
            this.getNode().item.applyDesignMetaData({
                $layout: {
                    $layoutType: "composite",
                    $layoutSubType: menuItem.$bind
                }
            });
            return false;
        }
        else {
            switch (menuItem.$item.$bind) {
                case "$addSection":
                case "$addBlock":
                    var parentNode = self.getNode();
                    var $category = (menuItem.$item.$bind == "$addBlock") ? "block" : "section";
                    var parentLevels = ($category == "block") ? ["article", "section", "block"] : ["article", "section"];
                    var isFieldChildren, newBox;
                    while (parentNode) {
                        if (parentLevels.indexOf(parentNode.$authoringLevel) >= 0) {
                            if (parentNode.$authoringLevel == $category) {
                                isFieldChildren = parentNode.parentNode.getChildrenLevel() == "field";
                                newBox = self.authorPage.addNewBox({
                                    parentBox: parentNode.parentNode.item,
                                    $category: $category,
                                    targetItem: parentNode.item,
                                    action: "insertAfter"
                                });
                            }
                            else {
                                isFieldChildren = parentNode.getChildrenLevel() == "field";
                                newBox = self.authorPage.addNewBox({
                                    parentBox: parentNode.item,
                                    $category: $category
                                });
                            }
                            break;
                        }
                        parentNode = parentNode.parentNode;
                    }
                    if (isFieldChildren) {
                        self.authorPage.moveContent({
                            sourceNode: newBox.authoringNode.parentNode,
                            targetItem: newBox
                        });
                    }
                    document.site.authorPage.notifyUpdate();
                    return false;
                case "$addField":
                    self.onAddFields(menuItem);
                    return false;
                case "$switchMode":
                    self.authorPage.targetPage.authoringNode.toggleLightMode(null, self.authorPage.$isLightMode = !self.authorPage.$isLightMode);
                    self._setSwitchMode();
                    return false;
            }
        }
    },
    loadBox: function(){
        var self = this;
        self.$item.$isAuthoringDisabled = true;
        var $isEditMode = self.getNode().item.getArticle().$isEditMode;
        var $properties = self.$prototype.$properties;
        if (self.isFieldItem()) {
            if (self.$prototype.$isTopLabelAlignment = $isEditMode) {
                $properties.$isRightTextLabelAlignment.$isDisabled = true;
            }
        }
        else {
            if (self.$prototype.$fieldsIsTopLabelAlignment = $isEditMode) {
                $properties.$fieldsIsRightTextLabelAlignment.$isDisabled = true;
            }
        }
        self._setBoxLayouType();
        
        /*var test = JSON.stringify(self.$prototype);
         var test = JSON.stringify(self.$item);*/
        RawPage.prototype.loadBox.call(this);
        self._addMenus();
        self._addLayoutsHeader();
        self.applyChange(self.getNode().item.$item);
        
        if (this.getNode().item.getAuthoringWidget) {
            if (this.fieldAuthoring = this.getNode().item.getAuthoringWidget()) {
                this.fieldAuthoring.$$container = this.layoutContent.items[1].$$body;
                this.page.initializeNewItem(this.fieldAuthoring, {}, this);
                this.fieldAuthoring.designedField = this.getNode().item;
                this.fieldAuthoring.$layoutOptions = document.site.$item.$layoutOptions.authoring;
                this.fieldAuthoring.$skin = this.fieldAuthoring.$layoutOptions.section;
                this.fieldAuthoring.loadBox(this.getNode().item.$item);
            }
        }
        
    },
    _addLink: function($bind, $value, $$container){
        return this.authorPage.loadNewItem($$container, {
            $bind: $bind,
            $category: "link",
            $skin: "s-aw-designer-link",
            $icon: {
                $mode: "icon",
                $path: "authoring/s-aw-"
            },
            $value: $value,
        }, this);
    },
    _addMenus: function(){
        var self = this;
        self._addLink("$switchMode", "switch-full", this.authorPage.$$switchModeSlot.empty());
        self._setSwitchMode();
        self.authorPage.$$panelActions.empty();
        ["$addSection", "$addBlock", "$addField"].forEach(function($bind){
            self._addLink($bind, $bind.substr(1).toLowerCase(), self.authorPage.$$panelActions);
        });
    },
    _addLayoutsHeader: function(){
        var self = this;
        self.authorPage.$$layoutSlot.empty();
        if (self.$prototype.$properties.$boxLayoutType) {
            var div = document.createElement("div");
            div.className = "s-aw-header-layouts-slot";
            var field = self.loadNewItem($(div).appendTo(self.authorPage.$$layoutSlot), {
                $bind: "$boxLayoutType",
                $icon: {
                    $inputMode: "icon",
                    $mode: "iconText",
                    $path: "authoring/s-layout-",
                    $width: "25px",
                    $height: "25px"
                },
                $format: "$combo",
                $choiceLayout: "3",
                $isTopLabelAlignment: false,
                $isEditMode: true,
                $skin: "s-aw-field",
                $isTitleHidden: true,
                $isLeftSpaceHidden: true
            }, self);
            field.$$item.attr("data-s-article", self.id);
            
            div = document.createElement("div");
            div.className = "s-aw-header-layouts-slot";
            field = self.loadNewItem($(div).appendTo(self.authorPage.$$layoutSlot), {
                $bind: "$allChildLayout",
                $skin: "s-aw-field",
                $icon: {
                    $inputMode: "icon",
                    $mode: "iconText",
                    $path: "authoring/s-layout-",
                    $width: "25px",
                    $height: "25px"
                },
                $format: "$combo",
                $choiceLayout: "1",
                $isTopLabelAlignment: false,
                $isEditMode: true,
                $isTitleHidden: true,
                $isLeftSpaceHidden: true
            }, self);
            field.$$item.attr("data-s-article", self.id);
            
            div = document.createElement("div");
            div.className = "s-aw-header-layouts-slot";
            div.setAttribute("data-s-article", self.id);
            var $subMenus = [];
            var $links = {};
            layoutSettings.getCompositeTypes().forEach(function($key){
                $subMenus.push({
                    $bind: $key
                });
                $links[$key] = {
                    $title: layoutSettings.getComposite($key).$title || self.localize.aw_layoutComposite
                };
            });
            self.loadNewItem($(div).appendTo(self.authorPage.$$layoutSlot), {
                "$category": "menus",
                "$title": self.localize.aw_selectTemplate,
                "$skin": "s-test-menus-icon",
                "$isBoxCollapsable": true,
                "$isPopupContent": true,
                "$titleIcon": {
                    "$value": "template",
                    "$path": "authoring/s-layout-"
                },
                "$itemIcon": {
                    "$mode": "iconText",
                    "$path": "authoring/s-layout-"
                },
                "$layout": {
                    "$items": $subMenus
                }
            }, self);
            self.applyChange({
                $links: $links
            });
        }
    },
    _setBoxLayouType: function(){
        var self = this;
        if (self.$prototype.$properties.$boxLayoutType) {
            var $enums = layoutSettings.$columns.map(function($width){
                var $parts = $width.split(",");
                return {
                    $value: $parts.join("-"),
                    $title: $parts.length + " " + self.localize.aw_layoutColumns
                };
            });
            $enums.unshift({
                $value: "100",
                $title: self.localize.aw_layoutColumn
            });
            self.$prototype.$properties.$boxLayoutType.$enum = $enums;
        }
    },
    dispose: function(){
        if (this.fieldAuthoring) {
            delete this.fieldAuthoring.designedField;
            delete this.fieldAuthoring;
        }
        if (this.$$popupPanel) {
            this.$$popupPanel.undelegate();
        }
        delete this.authorPage;
        delete this.localize;
        RawPage.prototype.dispose.call(this);
    }
});
