// syracuse-ui/lib/site/authoring/palette/pagePalette.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var ItemPalette = require("./itemPalette").ItemPalette;



function PagePalette(){
}

exports.PagePalette = helpers.defineClass(PagePalette, ItemPalette, {
    loadBox: function(){
        var self = this;
        self.$childCategory = "section";
        self.$prototype = {
            "$properties": {
                $isSeparatorsVisible: {
                    $type: "application/x-boolean",
                    $title: self.localize.aw_isSeparatorsVisible
                },
                $fieldsIsTitleHidden: {
                    $type: "application/x-boolean",
                    $title: self.localize.aw_isLabelHidden,
                    $isHidden: true
                }
            },
            $isSeparatorsVisible: false
        
        };
        self._formatFieldWidth(true);
        self._formatFieldLabelAlign(true);
        self._formatFieldTextLabelAlign(true);
        self._formatLayout();
        self._formatChildLayout();
        self.initializeNewItem(self, {
            $category: "section",
            $isSeparatorsVisible: true,
            $skin: "s-aw-designer-item",
            $layout: {
                $layoutType: "row",
                $fitToContent: true,
                $items: [{
                    $category: "section",
                    $title: self.localize.aw_appearance,
                    $layout: {
                        $items: [{
                            $bind: "$isSeparatorsVisible",
                            $skin: "s-aw-field",
                            $isEditMode: true,
                            $isLeftSpaceHidden: true
                        }]
                    }
                }, this._formatAllFieldsSection()]
            
            }
        });
        ItemPalette.prototype.loadBox.call(self);
    }
    
});
