// syracuse-ui/lib/site/authoring/palette/sectionBlockPalette.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var ItemPalette = require("./itemPalette").ItemPalette;


function SectionBlockPalette(){
}

exports.SectionBlockPalette = helpers.defineClass(SectionBlockPalette, ItemPalette, {
    loadBox: function(){
        this.$childCategory = this.authorPage.designedNode.$authoringLevel == "section" ? "block" : "field";
        this.$prototype = {
            "$properties": {
                $isHidden: {
                    $type: "application/x-boolean",
                    $title: this.localize.aw_isHidden
                },
                $title: {
                    $type: "application/x-string",
                    $title: this.localize.aw_boxTitle
                },
                $isMaximizable: {
                    $type: "application/x-boolean",
                    $title: this.localize.aw_isMaximizable
                },
                $isBoxCollapsable: {
                    $type: "application/x-boolean",
                    $title: this.localize.aw_isBoxCollapsable
                },
                $isTitleHidden: {
                    $type: "application/x-boolean",
                    $title: this.localize.aw_isTitleHidden
                },
                $isSeparatorsVisible: {
                    $type: "application/x-boolean",
                    $title: this.localize.aw_isSeparatorsVisible
                },
                $fieldsIsTitleHidden: {
                    $type: "application/x-boolean",
                    $title: this.localize.aw_isLabelHidden,
                    $isHidden: this.$childCategory == "block"
                }
            },
            $isSeparatorsVisible: false
        };
        this._formatFieldWidth(true);
        this._formatFieldLabelAlign(true);
        this._formatFieldTextLabelAlign(true);
        this._formatLayout();
        this._formatChildLayout();
        this.initializeNewItem(this, this._formatSectionBlock());
        ItemPalette.prototype.loadBox.call(this);
    }
});
