// etna-studio/lib/representations/x3js-class.js
"use strict";
exports.representation = {
	$entityName: "x3jsClass",
	$facets: {
		$details: {
			$layout: {
				$layoutType: "tabs",
				$items: [{
					$category: "section",
					$title: "General",
					$layout: {
						$items: [{
							$bind: "module"
						}, {
							$bind: "name"
						}, {
							$bind: "title"
						}, {
							$bind: "dictCode"
						}, {
							$bind: "status"
						}, {
							$bind: "comment"
						}, {
							$bind: "endpoint"
						}]
					}
				}, {
					$category: "section",
					$title: "Properties",
					$layout: {
						$items: [{
							$bind: "properties"
						}]
					}
				}, {
					$category: "section",
					$title: "Source",
					$layout: {
						$items: [{
							$bind: "source"
						}]
					}
				}]
			}
		},
		$edit: {
			$copy: "$details"
		},
		$query: {
			$layout: {
				$items: [{
					$category: "section",
					$layout: {
						$items: [{
							$category: "section",
							$layout: {
								$items: [{
									$bind: "$resources",
									$layout: {
										$items: [{
											$bind: "module"
										}, {
											$bind: "name"
										}, {
											$bind: "title"
										}, {
											$bind: "dictCode"
										}, {
											$bind: "status"
										}, {
											$bind: "comment"
										}, {
											$bind: "source"
										}, {
											$bind: "endpoint"
										}]
									}
								}]
							}
						}]
					}
				}]
			}
		}
	}
};
