// syracuse-ui/lib/links/navigation/sitemap/sitemap.js
"use strict";
var Scrollbar = require('syracuse-ui/lib/scroll/scrollbar').Scrollbar;

function _clickExpand(event) {
	var block = this.parent;
	block.dataset.isExpanded = !(block.dataset.isExpanded !== false);
	if (event.shiftKey) {
		var children = block.navParent.children;
		for (var ii = 0, jj = children.length; ii < jj; ii++) {
			var child = children[ii];
			if (!child.isMenuItem) {
				child.dataset.isExpanded = block.dataset.isExpanded;
				_expand(child);
			}
		}
	} else {
		_expand(block);
	}
	block.page.modulesScrollBar.resize();
	block.page.contentScrollBar.resize();
}

function _expand(block) {
	if (block.dataset.isExpanded) {
		syra_dom.hide(block.body, false);
		syra_button.setText(block.titleBtn, undefined, "node_expanded");
		block.titleBtn.link.syraTipTitle = syra_local.nvpCollapseSubModule;

		syra_button.disable(block.btnCollapseAll, false);
		syra_button.disable(block.btnExpandAll, true);

	} else {
		syra_dom.hide(block.body, true);
		syra_button.setText(block.titleBtn, undefined, "node_collapsed");
		block.titleBtn.link.syraTipTitle = syra_local.nvpExpandSubModule;

		syra_button.disable(block.btnCollapseAll, true);
		syra_button.disable(block.btnExpandAll, false);
	}
}

exports.page = {
	draw: function(page) {
		page.$skin = "s-nav-map";
		page.$skins = {
			module: "s-nav-record",
			submodule: "s-nav-map-submodule",
			nosubmodules: "s-nav-map-no-submodules",
			block: "s-nav-map-block",
			menu: "s-nav-map-link",
			nomenu: "s-nav-map-no-menu"
		};
		page.layoutSlot = syra_dom.div();
		page.domItem = syra_dom.div("", page.layoutSlot);
		page.body = syra_dom.div("", page.domItem);
		page.mapBody = syra_dom.div("s-nav-map-body", page.body);
		page.modulesBarSlot = syra_dom.div("s-nav-map-modules-slot", page.mapBody);
		page.modulesBar = syra_dom.ul("s-nav-map-modules", page.modulesBarSlot);
		page.modulesContentSlot = syra_dom.div("s-nav-map-modules-content-slot", page.mapBody);
		page.modulesScrollBar = new Scrollbar({
			container: page.modulesBarSlot,
			scrollview: page.modulesBar
		});
		page.modulesContent = syra_dom.ul("s-nav-map-modules-content", page.modulesContentSlot);
		page.contentScrollBar = new Scrollbar({
			container: page.modulesContentSlot,
			scrollview: page.modulesContent
		});

		page.resizeItem = function() {
			if (this.isMapLoaded && this.overPanel) {
				delete this.size;
				this.getPageSize();
				this.overPanel.resize();
			}
		};
		page.onNavMenuClick = function() {
			page.overPanel && page.overPanel.close();
		};
		page.modulesContent.addEventListener("scroll", page.modulesContent_onScroll = function(event) {
			for (var ii = 0, jj = page.children.length - 1; ii <= jj; ii++) {
				if (page.children[ii].scrollTopSelection > this.scrollTop || ii === jj) {

					var moduleIndex;
					if (page.contentScrollBar.isThumbAtBottom())
						moduleIndex = jj;
					else
					if (page.children[ii].scrollTopSelection > this.scrollTop)
						moduleIndex = Math.max((ii == jj + 1) ? ii : --ii, 0);
					else
						moduleIndex = Math.max((ii == jj) ? ii : --ii, 0);

					var module = page.children[moduleIndex];
					if (page.lastSelected != module) {
						module.scrolling = true;
						module.activate(true);
						module.scrolling = false;
					}
					break;
				}
			}
		});

		page.showBookmarks = true;

		syra_over.openModal(syra_site, {
			page: page,
			resize: function(dlg) {
				if (this.page.isMapLoaded) {
					this.page.modulesBarSlot.style.height = this.page.size.maxBarHeight + "px";
					this.page.modulesContentSlot.style.height = this.page.size.maxBarHeight + "px";

					var maxContentWidth = this.page.body.clientWidth - this.page.modulesBarSlot.clientWidth - 2;

					if (syra_site.$device == "tablet") {
						var modalHeight = syra_site.body.clientHeight - (syra_site.header.clientHeight + syra_site.footer.clientHeight);
						this.page.modulesBarSlot.style.height = modalHeight + "px";
						this.page.modulesContentSlot.style.height = modalHeight + "px";
						var width = syra_site.body.clientWidth;
						this.page.domItem.parentElement.className = "s-modal-page-container-tablet-sitemap";
						this.page.domItem.style.width = this.page.domItem.parentElement.style.maxWidth = width + "px";
						maxContentWidth = width - this.page.modulesBarSlot.clientWidth - 2;
						//this.page.modulesContent.style.width = maxContentWidth + "px";
						this.page.overPanel.hideCloseButton(true);
						this.page.overPanel.overlay.onclick = function() {
							dlg.close();
						};
					}

					this.page.modulesContentSlot.style.maxWidth = maxContentWidth + "px";
					this.page.modulesContent.style.width = maxContentWidth + "px";
					var height = this.page.modulesContentSlot.clientHeight;
					this.page.scrollTopSelection = height / 2;

					if (syra_site.$device == "tablet") {
						dlg.overSlot.style.top = syra_site.header.clientHeight + "px";
					} else {
						dlg.center();
					}

					for (var ii = 0, jj = this.page.children.length; ii < jj; ii++) {
						var module = this.page.children[ii];
						var offsetTop = module.content.offsetTop;
						if (offsetTop < height) {
							module.scrollTopSelection = offsetTop;
						} else {
							module.scrollTopSelection = offsetTop - this.page.scrollTopSelection;
						}
					}
					this.page.contentScrollBar.resize();
					this.page.modulesScrollBar.resize();
				}
			}
		});
	},
	dispose: function(page) {
		page.modulesScrollBar && page.modulesScrollBar.dispose();
		page.contentScrollBar && page.contentScrollBar.dispose();
		page.modulesContent && page.modulesContent_onScroll && page.modulesContent.removeEventListener("scroll", page.modulesContent_onScroll);
		delete page.modulesContent_onScroll;
	}
};

exports.module = {
	create: function(module) {
		syra_dom.span(module.$skin + "-title", module.slot).textContent = module.dataset.title;
		this._showSubModules(module);
	},
	activate: function(on, module) {
		if (on) {
			if (!module.scrolling) {
				module.content.parentNode.scrollTop = module.content.offsetTop;
			} else {
				module.page.scrollToItem(module.slot, module.page.modulesBar, true);
			}
		} else {

		}
	},
	_showSubModules: function(module) {
		var page = module.page;
		module.content = syra_dom.li("s-nav-map-module-content");
		module.titleBtn = syra_button.add({
			parent: module,
			slot: module.content,
			text: module.dataset.title,
			title: syra_local.nvpCollapseSubModule,
			css: "s-nav-map-module-content-title",
			fontIcon: "node_expanded",
			click: _clickExpand
		});

		module.body = syra_dom.ul("s-nav-map-submodules", module.content);
		module.addChildren();
		page.modulesContent.appendChild(module.content);
	}
};

exports.subModule = {
	create: function(sub) {
		sub.titleBtn = syra_button.add({
			parent: sub,
			slot: sub.slot,
			text: sub.dataset.title,
			title: syra_local.nvpCollapseSubModule,
			css: sub.$skin + "-title",
			fontIcon: "node_expanded",
			click: _clickExpand
		});

		sub.btnCollapseAll = syra_button.add({
			parent: sub,
			slot: sub.slot,
			text: syra_local.nvpCollapseAll,
			iconOnly: true,
			css: sub.$skin + "-node-expanded",
			fontIcon: "collapse_l",
			click: _clickExpand
		});

		sub.btnExpandAll = syra_button.add({
			parent: sub,
			slot: sub.slot,
			text: syra_local.nvpExpandAll,
			iconOnly: true,
			css: sub.$skin + "-node-expanded",
			fontIcon: "expand_l",
			click: _clickExpand
		});

		syra_button.disable(sub.btnCollapseAll, false);
		syra_button.disable(sub.btnExpandAll, true);

		sub.body = syra_dom.div(sub.$skin + "-body");
		sub.addChildren(sub.dataset.items);
		var nodes = sub.body.childNodes;
		if (nodes.length) {
			sub.cols = [];
			for (var ii = 0; ii < 3; ii++) {
				sub.cols.push(syra_dom.ul(sub.$skin + "-body-col"));
			}
			var ii = 0;
			while (nodes.length) {
				sub.cols[ii].appendChild(nodes[0]);
				if ((++ii) == 3) {
					ii = 0;
				}
			}
			for (var ii = 0; ii < 3; ii++) {
				sub.body.appendChild(sub.cols[ii]);
			}
		}
		sub.slot.appendChild(sub.body);
	},
	activate: function(on, block) {
		if (on) {
			block.navParent.lastSelected = block;
		} else {

		}
	}
};

exports.menuBlock = {
	create: function(block) {
		block.domItem = block.slot = syra_dom.li(block.$skin);
		block.titleBtn = syra_button.add({
			parent: block,
			slot: block.slot,
			text: block.dataset.title,
			title: syra_local.nvpCollapseSubModule,
			css: block.$skin + "-title",
			fontIcon: "node_expanded",
			click: _clickExpand
		});
		block.body = syra_dom.ul(block.$skin + "-body", block.slot);
	}
};
