// syracuse-ui/lib/fusion/field/collection/leftList.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var DesktopList = require("syracuse-ui/lib/field/array/list").DesktopList;
var sapUtil = require('syracuse-ui/lib/fusion/core/client/sap/util');
var jQuery = $;
//-----------------------------------------------
// GridSyra
//-----------------------------------------------
function Grid(){
    // Last selected Id returned by X3
    this._mySelSyraId = null;
    this._sortDataSet = null;
}

exports.Grid = helpers.defineClass(Grid, DesktopList, {
    createBuilder: function(){
        this.$item.$fitContainer = true;
        this.$item.$isContextMenuHidden = true;
        this.$item.$isQuickDesignerEnabled = false;
        this.$item.$isMenuRecordHidden = true
        DesktopList.prototype.createBuilder.call(this);
    },
    resize: function(height){
        this.onWindowResize();
    },
    applyMetaData: function(metaData){
        return false;
    },
    _applyPartialDelta: function (dataRecordSet, firstIndex, lastIndex) {
        _.each(dataRecordSet, function(meta){
            var syraId;
            if(meta.$isSelected != null && meta.$isSelected != undefined) {
                syraId = parseInt(meta.$index, 10) - 1;
                this._store.selector.select(syraId, meta.$isSelected)
                if (meta.$isSelected) {
                    self._mySelSyraId = syraId + "";
                }
            }
        }, this);
    },
    _isMulti: function(){
        return this.$item.$selectMode != "row";
    },
    beforeTrigger: function(evt, options){
        if (!options || !evt || !options.data) 
            return false;
        if (evt == "wdgt.list.selline") {
            if (sapUtil.Fusion.isFixtureMode(this)) {
                // We let Syra widget manage the event
                return false;
            }
            else if (!this._isMulti()) {
                // unselect current selected line to informe user that action has been taken into account
                var selLine = (parseInt(options.data.line, 10) - 1) + "";
                // check selLine to not unselect a line already selected
                if (this._mySelSyraId != null && selLine != this._mySelSyraId && this.dataset[this._mySelSyraId]) {
                    this._store.selector.select(this._mySelSyraId, false);
                }
                // Continue process - SendX3Action - Selection will be done by X3 reply
                return true;
            }
        }
        return true;
    },
    getQuickSelValues: function(){
        return this.filterCapability.filterRecord.fusionGetInputValues();
    },
    getXid: function(){
        return this.$item.$bind;
    },
    getIst: function(nl){
        return sapUtil.Fusion.makeIst(this.page.$fusionPageMeta.winModel.getWinId(), this.getXid(), nl === true && this._mySelSyraId != null ? parseInt(this._mySelSyraId, 10) : null);
    }
    /*,
     getDataSet: function(){
     if (this._sortDataSet) {
     return this._sortDataSet;
     }
     else {
     
     }
     }*/
});
