// syracuse-ui/lib/fusion/field/collection/listgrid.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var DesktopHgridBuilder = require('syracuse-ui/lib/desktop/field/collection/builder/hgridBuilder').Builder;
var DesktopList = require("syracuse-ui/lib/desktop/field/collection/list").DesktopList;
var sapUtil = require('syracuse-ui/lib/fusion/core/client/sap/util');
var jQuery = $;
//-----------------------------------------------
// GridSyra
//-----------------------------------------------
function Grid() {
}
exports.Grid = helpers.defineClass(Grid, DesktopList, {
    createBuilder: function() {
        this.$item.$isQuickDesignerEnabled = false;
        this.$item.$isMenuRecordHidden = true
        this.builder = new DesktopHgridBuilder();
        this.builder.list = this;
        this.builder.initialize();
    },
    resize: function(height) {
        if (this.$$item) this.$$item.height(height);
    },
    drawBox: function() {
        DesktopList.prototype.drawBox.call(this);
        this.getPage().fusionBar.addLeftList(this);
    },
    applyMetaData: function(metaData) {
        if (!metaData) return;
        var stts = metaData.stt, self = this;
        if (stts) {
            // stts [[x3id,x3Idx,gui],[x3id,x3Idx,gui]]
            var records = [];
            jQuery.each(stts, function(i, stt) {
                if (stt.length > 2) {
                    var x3Idx = stt[1], gui = sapUtil.Fusion.getState(stt[2]);
                    self.selectRecord(x3Idx - 1, gui.$isSelected)
                }
            })
        }
    }
});
