// syracuse-ui/lib/fusion/field/collection/listgrid.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var DesktopHgridBuilder = require('syracuse-ui/lib/desktop/field/collection/builder/hgridBuilder').Builder;
var DesktopList = require("syracuse-ui/lib/desktop/field/collection/list").DesktopList;
var sapUtil = require('syracuse-ui/lib/fusion/core/client/sap/util');
var jQuery = $;
//-----------------------------------------------
// GridSyra
//-----------------------------------------------
function Grid() {
	// Last selected Id returned by X3
    this._mySelectedId = null;
}
exports.Grid = helpers.defineClass(Grid, DesktopList, {
    createBuilder: function() {
        this.$item.$isQuickDesignerEnabled = false;
        this.$item.$isMenuRecordHidden = true
        this.builder = new DesktopHgridBuilder();
        this.builder.list = this;
        this.builder.initialize();
    },
    resize: function(height) {
        if (this.$$item) this.$$item.height(height);
    },
    drawBox: function() {
        DesktopList.prototype.drawBox.call(this);
        this.getPage().fusionBar.addLeftList(this);
    },
    applyMetaData: function(metaData) {
        if (!metaData) return;
        var stts = metaData.stt, self = this;
        if (stts) {
            // stts [[x3id,x3Idx,gui],[x3id,x3Idx,gui]]
            var records = [];
            this._mySelectedId = null;
            jQuery.each(stts, function(i, stt) {
                if (stt.length > 2) {
                    var syraId = parseInt(stt[1], 10) - 1, gui = sapUtil.Fusion.getState(stt[2]);
                    if (gui.$isSelected != null) {
                        self.selectRecord(syraId, gui.$isSelected)
                        if (gui.$isSelected) {
                            self._mySelectedId = syraId + "";
                        }
                    }
                }
            })
        }
    },
    _isMulti: function() {
        return this.recordSelector.$selectMode != "row";
    },
    beforeTrigger: function(evt, options) {
        if (!options || !evt) return false;
        if (evt == "wdgt.list.selline") {
            if (sapUtil.fixtureMode(this)) {
                // We let Syra widget manage the event
                if (options.data.doEvent) options.data.doEvent();
                return false;
            } else if (!this._isMulti()) {
                // unselect current selected line to informe user that action has been taken into account
                if (this._mySelectedId != null) {
                    this.selectRecord(this._mySelectedId, false);
                }
                // Continue process - SendX3Action - Selection will be done by X3 reply
                return true;
            }
        }
        return true;
    }
});
