// @sage/syracuse-react/src/navigation/sitemapEdit/page/menu.tsx
"use strict";
import * as React from 'react';
import { Link } from '../../../link/link';
import { SitemapMenu, setMenuDescription } from '../../sitemap/items';
import { SortableLine } from '../../../common/sortable/sortableLine';
import { getLocalString } from '../../../culture/localString';

interface IProps {
    menu: SitemapMenu,
    onStartSort?: (menu: SitemapMenu, handle: HTMLElement) => void,
    onDelete?: (menu: SitemapMenu) => void
}

interface IState {
    description?: string,
    bookmarked?: boolean,
    bookmarkVisible?: boolean
}

export class Menu extends React.Component<IProps, IState>{
    constructor(props: IProps, context: any) {
        super(props, context);
        this.state = {
            description: setMenuDescription(props.menu.dataset),
            bookmarked: false,
            bookmarkVisible: false
        };
    }
    handleClick = () => { }
    render() {
        return <SortableLine
            id={this.props.menu.$uuid}
            className="s_sitemapedit_menu"
            value={this.props.menu}
            onStartSort={this.props.onStartSort}>
            <Link
                className="s_sitemapedit_menu_link"
                title={this.props.menu.dataset.title}
                description={this.state.description}
                onClick={this.handleClick}
            />
            {this.props.onDelete &&
                <Link className="s_sitemapedit_icon_link"
                    icon="perm_delete"
                    title={getLocalString("sitemapedit_delete")}
                    value={this.props.menu}
                    onClick={this.props.onDelete} />}
        </SortableLine>;
    }
}
