// syracuse-collaboration/lib/entities/soap/client/soapStubOperation.js
"use strict"
var util = require("util");

exports.entity = {
	$properties: {
		name: {
			$title: "Name",
			$isMandatory: true,
			$isUnique: true,
//			$links: {
//				$details: {
//					$type: "application/json;vnd.sage=syracuse",
//					$url: "http://www.google.fr"
//				}
//			}
		},
		description: {
			$title: "Description",
		},
		tester: {
			$title: "Tester",
			$format: "$url"
		},
		soapAction: {
			$title: "Soap Action",
		},
		style: {
			$title: "Style",
		},
		location: {
			$title: "Location",
		},
		inputEncoding: {
			$title: "Input encoding",
		},
		inputName: {
			$title: "Input name",
			$isMandatory: true,
		},
		inputPrototype: {
			$title: "Input prototype",
			$isMandatory: true,
			$type: "binary",
			$storage: "db_file",
		},
		outputEncoding: {
			$title: "Output encoding",
		},
		outputName: {
			$title: "Output name",
			$isMandatory: true,
		},
		outputPrototype: {
			$title: "Output prototype",
			$isMandatory: true,
			$type: "binary",
			$storage: "db_file",
		},
	},
	$relations: {},
	$defaultOrder: [["name", true]]
}
