// etna/lib/engine/runtime/dbevaluate.ts
"use strict";

import * as bcd from './tbcd';
import * as db from './db';
const glob = require('streamline/lib/globals');
import { Parser } from '../parser';

import * as util from './util';
import * as basic from './basic';
import * as errors from './errors';

const X3Error = errors.X3Error;

function buildLiteral(node) {
	return basic.instructions.C(node.value);
}

function buildDecimal(node) {
	return basic.instructions.C(bcd.fromString(node.value));
}

var _buildExports;
var _buildRequire;

const builders = {
	STRING: buildLiteral,
	DECIMAL: buildDecimal,
	INTEGER: buildLiteral,
	IDENT(node) {
		return db.instructions.SQLV(node.value);
	},
	CLAS(node) {
		return db.instructions.SQLCV(node.letter, node.abbrev, node.field.toUpperCase());
	},
	OP(node) {
		return db.instructions["SQL" + node.op].apply(null, node.args.map(_build));
	},
	FCALL(node) {
		return db.instructions.SQLF(node.name, (node.args || []).map(_build));
	},
	EVALUE(node) {
		return db.instructions.SQLEVALUE(_buildRequire, _buildExports, _build(node.arg));
	},
	INDEX(node) {
		return db.instructions.SQLINDEX(_build(node.target), node.args.map(_build));
	},
	DATE(node) {
		return db.instructions.DATE(node.year, node.month, node.day);
	}
};

function _build(node) {
	const builder = builders[node.tag];

	if (!builder) throw new X3Error(200, "unsupported eval construct: " + node.tag);
	return builder(node);
}

export const functions = {
	EVALUE: { // parser needs it in this table but fn will never be called
		count: -1
	}
};

export const instructions = {
	EVALUE(req, ns, arg) {
		return function evalue$do(_: _) {
			const frame = glob.context.x3frame;
			const ctx = frame.whereContext;

			var argVal = arg(_);
			if (Array.isArray(argVal) && argVal[0] == 1) {
				var exp = "";
				for (var i = 0; i < argVal[2].length; i++) {
					exp += argVal[2][i];
				}
				argVal = exp;
			}
			argVal = util.x3ToString(argVal, true);
			try {
				var node = new Parser(argVal, "eval").parseExp(true);
				_buildRequire = req;
				_buildExports = ns;
				var fn = _build(node);
				return fn(_, ctx);
			} catch (e) {
				if (e.errn === 200) throw new X3Error(5, "Illegal character");
				else if (e.errn === 13) throw new X3Error(26, "Internal error");
				throw e;
			}
		};
	}
};
