// etna/lib/engine/runtime/evaluate.ts
"use strict";

import * as bcd from './tbcd';
/*
 * Eval
 */

const Parser = require("../parser").Parser;

var insts;

import * as util from './util';
import * as errors from './errors';
const X3Error = errors.X3Error;
const trace = require('syracuse-core').getTracer("etna.engine");

function buildLiteral(node) {
	return util.instructions.C(node.value);
}

function buildDecimal(node) {
	return util.instructions.C(bcd.fromString(node.value));
}

var _buildExports;
var _buildRequire;

const builders = {
	STRING: buildLiteral,
	DECIMAL: buildDecimal,
	INTEGER: buildLiteral,
	IDENT(node) {
		return insts.V(node.value);
	},
	PROP(node) {
		if (node.name.toUpperCase() !== "SNAPSHOT") {
			return insts.PROP(_build(node.target), node.name);
		} else {
			return insts.SNAP(_build(node.target));
		}
	},
	CLAS(node) {
		if (node.field && node.field.length) {
			return insts.CV(node.letter, node.abbrev, node.field.toUpperCase());
		} else {
			return insts.CLA(node.letter, node.abbrev);
		}
	},
	OP(node) {
		return insts[node.op].apply(null, node.args.map(_build));
	},
	FCALL(node) {
		return insts.F(node.name, (node.args || []).map(_build));
	},
	EVALUE(node) {
		return insts.EVALUE(_buildRequire, _buildExports, _build(node.arg));
	},
	SIGMA(node) {
		return insts.SIGMA(node.field, node.args.map(_build));
	},
	FUNC(node) {
		//dbg:console.error("FUNC node.from:",node.from)
		//dbg:console.error("FUNC node.name:",node.name)
		if (node.from) {
			return insts.CALL(insts.REQ(_buildRequire, _build(node.from)), node.name, (node.args || []).map(_build));
		} else {
			return insts.CALL(insts.C(_buildExports), _build(node.name), (node.args || []).map(_build));
		}
	},
	INDEX(node) {
		return insts.INDEX(_build(node.target), node.args.map(_build));
	},
	DATE(node) {
		return insts.DATE(node.year, node.month, node.day);
	},
	DIM(node) {
		var letter, field, rhs;
		for (var i = 0; i < node.args.length; i++) {
			switch (node.args[i].tag) {
				case "CLAS":
					letter = node.args[i].letter;
					field = node.args[i].field;
					break;
				case "IDENT":
					field = node.args[i].value;
					break;
				case "INTEGER":
					rhs = insts.C(node.args[i].value);
					break;
			}
		}
		if (field != null) return insts.DIM(letter || "", field || "", rhs);
		else return insts.DYNDIM.apply(null, node.args.map(_build));
	},
	TYPE(node) {
		if (node.field != null) return insts.TYPE(node.letter, node.field);
		else return insts.DYNTYPE(_build(node.exp));
	}
};

function _build(node) {
	const builder = builders[node.tag];

	if (!builder) {
		throw new X3Error(200, "unsupported eval construct: " + node.tag);
	}
	return builder(node);
}

export const functions = {
	EVALUE: { // parser needs it in this table but fn will never be called
		count: -1
	}
};

export const instructions = {
	EVALUE(req, ns, arg) {
		return function evalue$do(_) {
			insts = insts || require('./index').instructions;
			var argVal = arg(_);
			if (Array.isArray(argVal) && argVal[0] == 1) {
				var exp = "";
				for (var i = 0; i < argVal[2].length; i++) {
					exp += argVal[2][i];
				}
				argVal = exp;
			}
			argVal = util.x3ToString(argVal, true);
			if (!argVal.length) return ""; // Nothing to evaluate
			try {
				var node = new Parser(argVal, "eval").parseExp(true);
				_buildRequire = req;
				_buildExports = ns;
				var fn = _build(node);
				return fn(_);
			} catch (e) {
				trace.error && trace.error("EVALUE:" + argVal);
				if (e.errn === 200) throw new X3Error(5, "Illegal character:" + JSON.stringify(e));
				else if (e.errn === 13) throw new X3Error(26, "Internal error");
				throw e;
			}
		};
	}
};

export function compile(expression) {
	insts = insts || require('./index').instructions;
	const node = new Parser(expression, "eval").parseExp(true);

	return _build(node);
};

export function evalue(_, expression) {
	return compile(expression)(_);
};
