// etna/lib/engine/runtime/slots.ts
"use strict";

import { X3Error } from './errors';

export class Slot {
	_number: number;
	_next: number;
	_data: any;
}

export class Slots {
	_current: number;
	_slots: Slot[];
	_free: number;
	_size: number; // ??
	constructor() {
		this._current = 1;
		this._slots = [];
		this._free = 0;
		return this;
	}
	init() {
		return this;
	}
	size() {
		return this._size;
	}
	getSlot() {
		var s: Slot;
		if (this._free === 0) {
			s = {
				_number: this._current,
				_next: undefined,
				_data: undefined
			};
			this._slots[this._current] = s;
			this._current += 1;
		} else {
			s = this._slots[this._free];
			this._free = s._next;
			s._next = undefined;
		}
		return s._number;
	}
	freeSlot(n: number) {
		const s = this._slots[n];

		s._next = this._free;
		s._data = undefined;
		this._free = n;
	}
	get(n: number) {
		if (n < 1 || n >= this._current || this._slots[n]._next !== undefined) return undefined;
		return this._slots[n]._data;
	}
	set(n: number, data: any) {
		if (n < 1 || n >= this._current || this._slots[n]._next !== undefined) throw new X3Error(26, "Get invalid Slot" + this._current);
		this._slots[n]._data = data;
	}
};

;
