// hdb/lib-cov/protocol/part/StatementContext.js
/* automatically generated by JSCoverage - do not edit */
if (typeof _$jscoverage === 'undefined') _$jscoverage = {};
if (! _$jscoverage['protocol/part/StatementContext.js']) {
  _$jscoverage['protocol/part/StatementContext.js'] = [];
  _$jscoverage['protocol/part/StatementContext.js'][14] = 0;
  _$jscoverage['protocol/part/StatementContext.js'][16] = 0;
  _$jscoverage['protocol/part/StatementContext.js'][17] = 0;
  _$jscoverage['protocol/part/StatementContext.js'][18] = 0;
  _$jscoverage['protocol/part/StatementContext.js'][19] = 0;
  _$jscoverage['protocol/part/StatementContext.js'][20] = 0;
  _$jscoverage['protocol/part/StatementContext.js'][22] = 0;
  _$jscoverage['protocol/part/StatementContext.js'][24] = 0;
  _$jscoverage['protocol/part/StatementContext.js'][25] = 0;
  _$jscoverage['protocol/part/StatementContext.js'][28] = 0;
  _$jscoverage['protocol/part/StatementContext.js'][30] = 0;
  _$jscoverage['protocol/part/StatementContext.js'][33] = 0;
  _$jscoverage['protocol/part/StatementContext.js'][36] = 0;
  _$jscoverage['protocol/part/StatementContext.js'][37] = 0;
  _$jscoverage['protocol/part/StatementContext.js'][38] = 0;
  _$jscoverage['protocol/part/StatementContext.js'][40] = 0;
  _$jscoverage['protocol/part/StatementContext.js'][42] = 0;
  _$jscoverage['protocol/part/StatementContext.js'][48] = 0;
  _$jscoverage['protocol/part/StatementContext.js'][49] = 0;
  _$jscoverage['protocol/part/StatementContext.js'][50] = 0;
}
_$jscoverage['protocol/part/StatementContext.js'][14]++;
"use strict";
_$jscoverage['protocol/part/StatementContext.js'][16]++;
var util = require("../../util");
_$jscoverage['protocol/part/StatementContext.js'][17]++;
var AbstractOptions = require("./AbstractOptions");
_$jscoverage['protocol/part/StatementContext.js'][18]++;
var common = require("../common");
_$jscoverage['protocol/part/StatementContext.js'][19]++;
var PART_HEADER_LENGTH = common.PART_HEADER_LENGTH;
_$jscoverage['protocol/part/StatementContext.js'][20]++;
module.exports = StatementContext;
_$jscoverage['protocol/part/StatementContext.js'][22]++;
util.inherits(StatementContext, AbstractOptions);
_$jscoverage['protocol/part/StatementContext.js'][24]++;
function StatementContext() {
  _$jscoverage['protocol/part/StatementContext.js'][25]++;
  AbstractOptions.call(this);
  _$jscoverage['protocol/part/StatementContext.js'][28]++;
  this.statementSequenceInfo = undefined;
  _$jscoverage['protocol/part/StatementContext.js'][30]++;
  this.serverExecutionTime = 0;
}
_$jscoverage['protocol/part/StatementContext.js'][33]++;
Object.defineProperties(StatementContext.prototype, {size: {get: (function getSize() {
  _$jscoverage['protocol/part/StatementContext.js'][36]++;
  var statementSequenceInfoLength;
  _$jscoverage['protocol/part/StatementContext.js'][37]++;
  if (this.statementSequenceInfo) {
    _$jscoverage['protocol/part/StatementContext.js'][38]++;
    statementSequenceInfoLength = this.statementSequenceInfo.length;
  }
  else {
    _$jscoverage['protocol/part/StatementContext.js'][40]++;
    statementSequenceInfoLength = 10;
  }
  _$jscoverage['protocol/part/StatementContext.js'][42]++;
  return PART_HEADER_LENGTH + util.alignLength(4 + statementSequenceInfoLength, 8);
})}});
_$jscoverage['protocol/part/StatementContext.js'][48]++;
StatementContext.prototype.PROPERTY_NAMES = common.StatementContextName;
_$jscoverage['protocol/part/StatementContext.js'][49]++;
StatementContext.prototype.TYPES = common.StatementContextType;
_$jscoverage['protocol/part/StatementContext.js'][50]++;
StatementContext.prototype.KEYS = [common.StatementContext.STATEMENT_SEQUENCE_INFO];
_$jscoverage['protocol/part/StatementContext.js'].source = ["// Copyright 2013 SAP AG.","//","// Licensed under the Apache License, Version 2.0 (the \"License\");","// you may not use this file except in compliance with the License.","// You may obtain a copy of the License at","//","// http: //www.apache.org/licenses/LICENSE-2.0","//","// Unless required by applicable law or agreed to in writing,","// software distributed under the License is distributed on an","// \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,","// either express or implied. See the License for the specific","// language governing permissions and limitations under the License.","'use strict';","","var util = require('../../util');","var AbstractOptions = require('./AbstractOptions');","var common = require('../common');","var PART_HEADER_LENGTH = common.PART_HEADER_LENGTH;","module.exports = StatementContext;","","util.inherits(StatementContext, AbstractOptions);","","function StatementContext() {","  AbstractOptions.call(this);","","  // Information on the statement sequence within the transaction","  this.statementSequenceInfo = undefined;","  // Time for statement execution on the server in nano seconds","  this.serverExecutionTime = 0;","}","","Object.defineProperties(StatementContext.prototype, {","  size: {","    get: function getSize() {","      var statementSequenceInfoLength;","      if (this.statementSequenceInfo) {","        statementSequenceInfoLength = this.statementSequenceInfo.length;","      } else {","        statementSequenceInfoLength = 10;","      }","      return PART_HEADER_LENGTH + util.alignLength(4 +","        statementSequenceInfoLength, 8);","    }","  }","});","","StatementContext.prototype.PROPERTY_NAMES = common.StatementContextName;","StatementContext.prototype.TYPES = common.StatementContextType;","StatementContext.prototype.KEYS = [","  common.StatementContext.STATEMENT_SEQUENCE_INFO","];"];
