// syracuse-ui/lib/site/url/history.js
"use strict";
var globals = require('streamline/lib/globals');

var _updateHitory;

exports.temphackWorkbook = function() {
	_updateHitory = true;
	window.history.forward();
};

exports.start = function() {
	//load representation of url
	exports.reload();
	//listen history change
	window[window.history.pushState ? "onpopstate" : "onhashchange"] = function(event) {
		if (_updateHitory) {
			_updateHitory = false;
		} else {
			exports.reload();
		}
	};
};

exports.update = function(url) {
	var param = document.site.urlMaker.setUrlParameter(url);
	if (window.history.pushState) {
		window.history.replaceState("$updateHitory", param, param);
	} else {
		_updateHitory = true;
		window.location.hash = "#" + encodeURIComponent(param);
	}
};

exports.load = function(url, target) {
	if (target == "blank") {
		window.open(url, "_blank");
	} else {
		if (window.history.pushState) {
			window.history.pushState("", url, url);
			document.controller.changeMainPage(url);
		} else {
			if (url.indexOf("#") >= 0) {
				window.location.assign(url);
				document.controller.changeMainPage(url);
			} else {
				window.location.hash = "#" + encodeURIComponent(url);
			}
		}
	}
};


exports.reload = function() {
	// handle LTR / RTL switch
	var lp = globals.context && globals.context.localePreferences;
	var rtl = lp && lp.code && /^(ar|iw)/.test(lp.code);
	var href = document.location.href;
	var qm = href.indexOf('?');
	if (rtl && !/rtl=true/.test(href))
		href += (qm < 0 ? '?' : '&') + 'rtl=true';
	else
	if (!rtl)
		href = href.replace(/(\?|\&)rtl=true/, '');
	// reload page if writing direction changes
	if (href != document.location.href)
		document.location.href = href;
	else
		document.controller.changeMainPage(href);
};

exports.updateCurrentEndpoint = function(newEndpoint) {
	var url = document.site.urlMaker.getUrlParameter(document.location.href);
	exports.update(document.site.urlMaker.replaceEndpoint(url, newEndpoint));
	exports.reload();
};
