// @sage/syracuse-ui/lib/filter/operators.js
"use strict";
var _helpers = require('@sage/syracuse-core').helpers;
var List = require('syracuse-ui/lib/scroll/list').List;

function _clickOperator(event) {
	var field = this.parent;
	var old = field.filter_selectedOperator;
	exports.setFieldOperator(field, undefined, this.operator);
	if (field.filter_selectedOperator != old) {
		if (field.articleParent.list && (field.filter_selectedOperator == "between" || old == "between")) {
			field.articleParent.list.resizeItem();
		}
	}
	field.page.filter_popup && field.page.filter_popup.close();
	field.page.externalAdapter.onFilterEvent({
		operator: field.filter_selectedOperator,
		list: field.articleParent.list,
		filter: field,
		event: event,
		type: "operatorchange",
		doEvent: function() {
			var record = field.articleParent;
			var value = field.getInputValue();
			if (field.filter_selectedOperator != "none") {
				if (value != null && value != "") {
					if (field.filter_selectedOperator == "empty" || field.filter_selectedOperator == "notempty") {
						syra_filter.setCriteria(field, {
							value: null,
							operator: field.filter_selectedOperator
						});
					}
					record.triggerFilter();
				} else {
					if (field.filter_selectedOperator == "empty" || field.filter_selectedOperator == "notempty") {
						record.triggerFilter();
					}
				}
			} else {
				if (value != null && value != "") {
					syra_filter.setCriteria(field, {
						value: null,
						operator: field.filter_selectedOperator
					});
					syra_form.update(field, null);
				} else {
					record.triggerFilter();
				}
			}
		}
	});
}

function _clickShowList(event) {
	var field = this.parent;
	var record = field.articleParent;
	record.page.externalAdapter.onFilterEvent({
		operator: field.filter_selectedOperator,
		list: record.list,
		event: event,
		type: "operatorclick",
		doEvent: function() {
			if (field.page.filter_popup) {
				field.page.filter_popup.close();
			} else {
				var btns = [];
				(field.filter_list = new List({
					css: "s-list-default",
					scroll: true
				})).scrollSlot.syraItem = field.id;
				exports.assignToField(field);
				for (var ii = 0, jj = field.filter_operators.length; ii < jj; ii++) {
					var op = field.filter_operators[ii];
					var btn = field.filter_list.addButton({
						parent: field,
						text: syra_local["flFilter_" + op],
						fontIcon: "filter_" + op,
						click: _clickOperator,
						operator: op
					});
					if (field.filter_selectedOperator == op) {
						btn.link.className += " s-selected";
					}
					btns.push(btn);
				}
				field.page.filter_popup = syra_popup.open({
					slot: field.filter_list.scrollSlot,
					of: field.layoutSlot,
					onResize: function(maxHeight) {
						if (field.filter_list) {
							field.filter_list.resize(maxHeight);
						}
					},
					close: function() {
						field.filter_list && field.filter_list.dispose();
						delete field.filter_list;
						if (field.page) {
							field.page.filter_popup = null;
						}
					}
				});
				syra_quality.listFilter(field, btns);
			}
		}
	});
}

function _getFieldOperators(field) {
	var $type = (field.$filterField && field.$filterField.$type) || field.$field.$type;
	if (field.page.isFusionPage) {
		switch ($type) {
			case "application/x-date":
			case "application/x-time":
			case "application/x-datetime":
				return ["none", "ge", "gt", "le", "lt", "eq", "ne"];
			case "application/x-string":
				return ["none", "like", "like_s", "like_e", "ge", "gt", "le", "lt", "eq", "ne"];
			case "application/x-choice":
			case "application/x-real":
			case "application/x-decimal":
			case "application/x-integer":
			case "application/x-quantity":
				return ["none", "eq", "ge", "gt", "le", "lt", "ne"];
			default:
				return ["none", "eq", "ne"];
		}
	} else {
		switch ($type) {
			case "application/x-date":
			case "application/x-time":
			case "application/x-datetime":
				var operators = ["none", "ge", "gt", "lt", "le", "eq", "ne"];
				if (field.$field.$type != "application/x-reference") {
					// operators.push("between");
				}
				return operators;
			case "application/x-string":
				return ["none", "like", "like_s", "like_e", "ge", "gt", "le", "lt", "eq", "ne"];
			case "application/x-binary":
			case "application/x-document":
			case "image":
				return ["none", "empty", "notempty"];
			case "application/x-boolean":
				return ["none", "eq", "ne"];
			case "application/x-decimal":
			case "application/x-integer":
			case "application/x-quantity":
				if (field.$field.$type != "application/x-reference") {
					return ["none", "eq", "ge", "gt", "le", "lt", "between", "ne"];
				}
				return ["none", "eq", "ne"];
		}
		return ["none", "eq", "ne"];
	}
}

exports.assignToField = function(field) {
	return field.filter_operators || (field.filter_operators = _getFieldOperators(field));
};

exports.getDefault = function(field) {
	return exports.assignToField(field)[1];
};

exports.addPicker = function(field, css) {
	field.filter_picker = syra_button.add({
		parent: field,
		text: syra_local.flfilter_choiceOperator,
		css: css,
		iconOnly: true,
		fontIcon: "filter_none",
		click: _clickShowList
	});
	field.filter_picker.link.setAttribute("data-s-list-filter-op", field.$item.$bind);
	field.filter_picker.link.removeAttribute("href");
	field._dataValue.insertBefore(field.filter_picker.link, field._dataValue.firstChild);
};

exports.setFieldOperator = function(field, value, operator) {
	if (operator === undefined) {
		operator = field.filter_selectedOperator;
		if (value == null || value == "") {
			operator = "none";
		} else {
			if (field.$field && operator == "none") {
				operator = exports.getDefault(field);
			}
		}
	}
	syra_dom.toggleClass(field.filter_slot, field.articleParent.filter_criteria_css + "-active", operator != "none");
	syra_button.setText(field.filter_picker, "", "filter_" + (field.filter_selectedOperator = operator));
	if (operator == "between") {
		if (!field.highField) {
			var $bind = field.$item.$bind;
			field.articleParent.$prototype.$properties[$bind + "$high"] = _helpers.object.clone(field.articleParent.$prototype.$properties[$bind], true);
			var $item = _helpers.object.clone(field.$item, true);
			$item.$bind = $bind + "$high";
			var slot = syra_dom.div("s-filter-between-right", field.filter_slot);
			field.highField = field.page.addItem(slot, $item, field.articleParent);
			syra_dom.toggleClass(field._dataValue, "s-filter-between-left", true);
		}
	} else {
		if (field.highField) {
			syra_item.remove(field.highField, true, true);
			delete field.articleParent.$prototype.$properties[field.$item.$bind + "$high"];
			delete field.highField;
			syra_dom.toggleClass(field._dataValue, "s-filter-between-left", false);
		}
	}
};
