// syracuse-tablet/html/js/controls/layout/layoutHub.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var Base = require('syracuse-tablet/html/js/controls/layout/layoutBase').Klass;

/**
 * Top level layout node that must only contain layoutHubGroups as children
 * This "hub" expands horizontally and only uses the available vertical space
 * Use this to define pages which only scroll horizontally
 */
var _Klass = utils.defineClass(
	function(controller, type, article, opts) {
		var self = this;
		Base.call(self, controller, type, article, opts);
		/*
		 * CSS-DEPENDEND: Needs to be adjusted on major size or page layout changes
		 */
		self.pageContentHeightSubstraction = 200;
		self.minLandscapeHeight = 550;
	}, Base, {

		destroy: function() {
			Base.prototype.destroy.call(this);
		},

		buildHtml: function($$parent, controllerDao, buildOptions) {
			var self = this;
			buildOptions = buildOptions || {};
			buildOptions.noTitle = true;
			Base.prototype.buildHtml.call(self, $$parent, controllerDao, buildOptions);
			self.$$elmt.addClass("s-m-hub");
			self.children.forEach(function(child) {
				child.buildHtml(self.$$elmt, controllerDao);
			});
		},
		/**
		 * Resize handler called by window.resize
		 */
		onResize: function(evt) {
			var self = this;
			var w = window.innerWidth - self.$$elmt.offset().left;
			var h = window.innerHeight;
			var orientation = (w > h && h >= self.minLandscapeHeight) ? "landscape" : "portrait";
			h = h - self.pageContentHeightSubstraction;
			self.$$elmt.height(h);
			var posX = 0;
			var posY = 0;
			var opts;
			this.children.forEach(function(child) {
				opts = {
					posX: posX,
					posY: posY,
					maxHeight: h,
					maxWidth: w,
					orientation: orientation
				};
				child.onResize(evt, opts);
				if (orientation === "landscape") {
					posX += child.getWidth() + 10;
				} else {
					posY += child.getHeight() + 10;
				}
			});
		}
	});

exports.Klass = _Klass;
