// syracuse-tablet/html/js/controls/layout/layoutHub.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var globals = require('syracuse-tablet/html/js/helpers/globals');
var Base = require('syracuse-tablet/html/js/controls/layout/layoutBase').Klass;
var layoutHubGroup = require('syracuse-tablet/html/js/controls/layout/layoutHubGroup');
var uiRect = require('syracuse-tablet/html/js/ui/rect');

/**
 * Top level layout node that must only contain layoutHubGroups as children
 * This "hub" expands horizontally and only uses the available vertical space
 * Use this to define pages which only scroll horizontally
 */
var _Klass = utils.defineClass(
	function LayoutHub(controller, type, article, opts) {
		var self = this;
		Base.call(self, controller, type, article, opts);
	}, Base, {

		destroy: function() {
			Base.prototype.destroy.call(this);
		},

		buildHtml: function($$parent, controllerDao, buildOptions) {
			var self = this;
			buildOptions = buildOptions || {};
			buildOptions.noTitle = true;
			Base.prototype.buildHtml.call(self, $$parent, controllerDao, buildOptions);
			self.$$elmt.addClass("s-m-hub");
			self.$$scrollElmt = $('<div style="position:relative; height:100%; width:100%">').appendTo(self.$$elmt);
			self.children.forEach(function(child) {
				child.buildHtml(self.$$scrollElmt, controllerDao, buildOptions);
			});
		},
		/**
		 * Resize handler called by window.resize
		 */
		onResize: function() {
			var self = this;
			/**
			 * Returns the visible rectangle of main page content
			 *
			 */
			var parentRect = self.controller.getMainContentRect();
			var pos = self.$$elmt.position();
			var w = parentRect.width;
			var h = parentRect.height - pos.top;
			var rel = w / h;
			var orientation = (rel > 1.2 && h >= layoutHubGroup.minLandscapeHeigth) ? "landscape" : "portrait";
			self.$$elmt.height(h);
			var posX = 0;
			var posY = 0;
			var opts;
			var width = 0;
			self.children.forEach(function(child) {
				opts = {
					posX: posX,
					posY: posY,
					maxHeight: h,
					maxWidth: w,
					orientation: orientation
				};
				child.onResize(opts);
				if (orientation === "landscape") {
					posX += child.getWidth() + 10;
				} else {
					posY += child.getHeight() + 10;
				}
				width += child.$$elmt.outerWidth();
			});
			// Used by scroller
			var o = self.$$elmt.offset();
			self.insideRect = uiRect.newRect(o.top, o.left, self.$$elmt.height(), width);
			self.$$scrollElmt.width(width);
			self.rect = uiRect.elmtRect(self.$$elmt);
		},
		getScrollViewRect: function() {
			var viewRect = uiRect.elmtRect(this.$$elmt, "outer");
			var scrollRect = uiRect.elmtRect(this.$$scrollElmt, "outer");
			return viewRect.contains(scrollRect) ? null : viewRect;
		},
		getScrollElement: function() {
			return this.$$scrollElmt;
		}
	});

exports.Klass = _Klass;
