// syracuse-ui/lib/tests/trackers/sample.js
"use strict";
exports.$title = "Samples";
exports.trackers = [{
	location: "/sdata/$trackers('63023233-9f27-41bb-9749-2331a99596ea')",
	uuid: "record0",
	$state: "Job started",
	phaseDetail: "print",
	progress: 25,
	elapsedSeconds: 10,
	remainingSeconds: 50,
	startDate: 1428699906078,
	title: "Pays",
	kind: "print",
	$links: {
		$download: {
			$title: "Download"
		}
	},
	$diagnoses: [0, 1, 2, 3, 4].map(function(ii) {
		return {
			$message: "number " + ii,
			$severity: "error"
		};
	})
}, {
	uuid: "record1",
	$state: "Job started",
	phaseDetail: "printing",
	progress: 50,
	elapsedSeconds: 10,
	remainingSeconds: 50,
	startDate: 1428699906078,
	title: "Compagnies",
	kind: "print",
	$diagnoses: [{
		$message: "This is a test diagnose : warning",
		$severity: "warning",
		$stackTrace: "TypeError: Object 1115 has no method 'replace' at C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:18:39 at Array.forEach (native) at change (C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:16:20) at C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:20:41 at Array.forEach (native) at change (C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:16:20) at C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:20:41 at Array.forEach (native) at change (C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:16:20) at C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:20:41 at Array.forEach (native) at change (C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:16:20) at C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:20:41 at Array.forEach (native) at change (C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:16:20) at C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:20:41 at Array.forEach (native) at change (C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:16:20) at C:\Syracuse\Syracuse\node_modules\syracuse-collaboration\lib\entities\page\pageData._js:20:41 at Array.forEach (native) "
	}],
	$links: {
		$cancel: {
			$title: "Cancel"
		}
	}
}, {
	uuid: "record2",
	$state: "Job started",
	phaseDetail: "updating index",
	progress: 50,
	elapsedSeconds: 10,
	remainingSeconds: 50,
	startDate: 1428699906078,
	title: "Index update",
	kind: "operation",
	$links: {
		$abort: {
			$title: "Abort"
		}
	}
}];
