// syracuse-ui/lib/tests/trackers/trackers.js
"use strict";
var _tests = [require("syracuse-ui/lib/tests/trackers/searchIndex")];

var _sample = require("syracuse-ui/lib/tests/trackers/sample");

exports.load = function(page) {
	page.$prototype.$properties.mode = {
		$title: "mode",
		$type: "application/x-choice",
		$value: {
			$type: "application/x-string",
			$enum: [{
				$value: "auto",
				$title: "auto"
			}, {
				$value: "manual",
				$title: "manual"
			}]
		}
	};

	page.loadNewItem(page.leftCol, {
		$isEditMode: true,
		$isTopLabelAlignment: false,
		$bind: "mode"
	});
	page.applyChange({
		mode: "auto"
	});

	for (var ii = 0, jj = _tests.length; ii < jj; ii++) {
		page.leftCol.appendChild(syra_menus.addTextButton(_tests[ii].$title + " ( " + _tests[ii].trackers.length + ")", "s-uitest-mn", "onTrackerClick")).syraValue = ii;
	}

	page.rightCol.appendChild(syra_menus.addTextButton(_sample.$title, "s-uitest-mn", "onTrackerSampleClick"));

	page.setTitle("Trackers");
	page.notifyDataChange = function(field, value) {

	};
	page.onTrackerClick = function(event, btn) {
		var test = _tests[btn.syraValue];
		if (!btn.syraStep) {
			btn.syraIntervalValue = btn.syraStep = 0;
			var data = test.trackers[btn.syraStep];
			data.isTestCase = true;
			test.tracker = syra_trackers.addTracker(data);
		}
		if (this.boundFields.mode[0].getDataValue() == "auto") {
			btn.syraStep++;
			btn.syraInterval = setInterval(function() {
				test.tracker.applyChange(test.trackers[btn.syraStep++], btn.syraStep == test.trackers.length);
				if (btn.syraStep == test.trackers.length) {
					clearInterval(btn.syraInterval);
					test.tracker = btn.syraStep = btn.syraInterval = 0;
				}
			}, 800);
		} else {
			syra_menus.setButtonTitle(btn, "Stop " + test.$title + " ( " + (btn.syraStep + 1) + "/" + test.trackers.length + ")");
			if (btn.syraStep == 0) {
				btn.syraStep++;
				return;
			}
			test.tracker.applyChange(test.trackers[btn.syraStep++], btn.syraStep == test.trackers.length);
			if (btn.syraStep == test.trackers.length) {
				test.tracker = btn.syraStep = 0;
				syra_menus.setButtonTitle(btn, test.$title + " ( " + test.trackers.length + ")");
			}
		}


	};
	page.onTrackerSampleClick = function(event, btn) {
		btn.syraStep = 0;
		btn.syraInterval = setInterval(function() {
			var data = _sample.trackers[btn.syraStep++];
			data.isTestCase = true;
			syra_trackers.addTracker(data);
			if (btn.syraStep == _sample.trackers.length) {
				clearInterval(btn.syraInterval);
				btn.syraStep = btn.syraInterval = 0;
			}
		}, 800);
	};
};
