// syracuse-ui/lib/desktop/field/collection/authoring/fieldsSelector.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');


function FieldsSelector(){

}

exports.FieldsSelector = helpers.defineClass(FieldsSelector, null, {
    load: function(list, $$container){
        var self = this;
        self.list = list;
        self.$$item = $("<table class='s-list-design-fields'/>").appendTo($$container);
        self.$$table = $("<table class='s-list-design-fields-table'/>");
        var $$rowHeader = $("<tr/>");
        var $rowBody = $("<tr/>");
        self.cols = [];
        [{
            $title: "Available",
            $css: "available"
        }, {
            $css: "arrows"
        }, {
            $title: "Display",
            $css: "available"
        }].forEach(function($col){
            var css = "s-list-design-fields-" + $col.$css;
            self.cols.push({
                $col: $col,
                $fullCss: css,
                $$th: $("<th class='s-list-design-fields-th'/>").addClass(css).appendTo($$rowHeader).text($col.$title || ""),
                $$td: $("<td class='s-list-design-fields-td'/>").addClass(css).appendTo($rowBody)
            });
        })
        self.$$table.append($("<thead/>").append($$rowHeader)).append($("<tbody/>").append($rowBody)).appendTo(self.$$item);
        self.appendArrows();
        self.appendDisplay();
    },
    appendDisplay: function(){
        var self = this;
        var col = self.cols[2];
        col.$$tdBody = $("<div class='s-list-design-fields-td-body'/>").appendTo(col.$$td);
        col.$$ul = $("<ul class='s-list-design-fields-ul'/>").addClass(self.$fullCss).appendTo(col.$$tdBody);
        self.list.$item.$layout.$items.forEach(function($item){
            var $field = self.list.$fields[$item.$bind];
            var $$li = $("<li class='s-list-design-fields-li'/>").appendTo(col.$$ul).attr("data-s-bind", $item.$bind).text($field.$title);
            $$li[0].style.width = $$li.width() + "px";
        });
        col.$$ul.sortable({
            containment: self.$$table,
            revert: true
        });
        col.$$ul.disableSelection();
    },
    appendArrows: function(){
        var self = this;
        var link = document.createElement("a");
        link.className = "s-list-design-fields-select";
        self.$$select = $(link).bind("click", function(){
            return false;
        });
        
        link = document.createElement("a");
        link.className = "s-list-design-fields-unselect";
        self.$$unselect = $(link).bind("click", function(){
            return false;
        });
        
        self.cols[1].$$td.append(this.$$select).append(this.$$unselect);
    },
    dispose: function(){
        delete this.list;
        if (this.$$select) {
            this.$$select.unbind();
            this.$$unselect.unbind();
        }
    }
});
