// syracuse-ui/lib/desktop/field/collection/authoring/fieldsSelector.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');


function FieldsSelector(){

}

exports.FieldsSelector = helpers.defineClass(FieldsSelector, null, {
    load: function(list, $$container){
        var self = this;
        self.list = list;
        self.$$item = $("<table class='s-list-design-fields'/>").appendTo($$container);
        self.$$table = $("<table class='s-list-design-fields-table'/>");
        var $$rowHeader = $(document.createElement("tr"));
        var $rowBody = $(document.createElement("tr"));
        self.availableList = {
            $btnCss: "s-list-design-fields-add",
            $isHiddenItem: true,
            $$th: $("<th class='s-list-design-fields-items-th'/>").appendTo($$rowHeader).text("Available"),
            $$td: $("<td class='s-list-design-fields-items-td'/>").appendTo($rowBody)
        };
        
        self.arrows = {
            $fullCss: "s-list-design-fields-items",
            $$th: $("<th class='s-list-design-fields-arrows-th'/>").appendTo($$rowHeader),
            $$td: $("<td class='s-list-design-fields-arrows-td'/>").appendTo($rowBody)
        };
        self.displayList = {
            $btnCss: "s-list-design-fields-del",
            $isHiddenItem: false,
            $$th: $("<th class='s-list-design-fields-items-th'/>").appendTo($$rowHeader).text("Display"),
            $$td: $("<td class='s-list-design-fields-items-td'/>").appendTo($rowBody)
        };
        self.$$table.append($$rowHeader).append($rowBody).appendTo(self.$$item);
        
        var link = document.createElement("a");
        link.className = "s-list-design-fields-select";
        self.$$select = $(link).appendTo(self.arrows.$$td);
        link = document.createElement("a");
        link.className = "s-list-design-fields-unselect";
        self.$$unselect = $(link).appendTo(self.arrows.$$td);
        
        self._fillList(self.availableList);
        self._fillList(self.displayList);
        
        self.$$item.delegate("a.s-list-design-fields-del,a.s-list-design-fields-add", "click", function(){
            var hideColumn = this.className.indexOf("s-list-design-fields-del") >= 0;
            var $bind = $(this).closest("[data-s-bind]").attr("data-s-bind");
            setTimeout(function(){
                self._moveItem($bind, hideColumn, true);
            });
            return false;
        }).delegate(".s-list-design-fields-li", "click", function(){
            $(this).toggleClass("s-list-design-fields-selected");
            return false;
        }).delegate(".s-list-design-fields-li", "dblclick", function(){
            var $bind = $(this).attr("data-s-bind");
            var hideColumn = ($(this).closest("td").index() == 2);
            setTimeout(function(){
                self._moveItem($bind, hideColumn, true);
            });
            return false;
        }).delegate("a.s-list-design-fields-select,a.s-list-design-fields-unselect", "click", function(){
            var hideColumn = this.className.indexOf("unselect") > 0;
            var sourceList = hideColumn ? self.displayList : self.availableList;
            var $$selectedItems = sourceList.$$ul.children(".s-list-design-fields-selected");
            if ($$selectedItems.length == 0) {
                document.site.showMessage({
                    $title: "Invalidate authoring action",
                    $message: "No selected field"
                });
            }
            else {
                if (hideColumn && $$selectedItems.length == self.displayList.$$ul.children().length) {
                    document.site.showMessage({
                        $title: "Invalidate authoring action",
                        $message: "At leat one field has to be displayed"
                    });
                }
                else {
                    self._moveItems($$selectedItems, hideColumn);
                }
                
            }
            return false;
        }).delegate("a.s-list-design-fields-select-all,a.s-list-design-fields-unselect-all", "click", function(){
            var $$list = ($(this).closest("th").index() == 2) ? self.displayList : self.availableList;
            $$list.$$ul.children("li").toggleClass("s-list-design-fields-selected", this.className.indexOf("unselect") < 0);
            return false;
        });
    },
    _moveItems: function($$selectedItems, $isHidden){
        var metadata = {};
        for (var ii = 0; ii < $$selectedItems.length; ii++) {
            var $bind = $$selectedItems[ii].getAttribute("data-s-bind");
            metadata[$bind] = {
                $isHidden: $isHidden
            };
            this._moveItem($bind, $isHidden);
        }
        this.list.applyDesignMetaData({
            $item: {
                $: metadata
            }
        }, true);
    },
    _moveItem: function($bind, $isHidden, notify){
        if (notify) {
            if ($isHidden && this.displayList.$$ul.children().length <= 1) {
                document.site.showMessage({
                    $title: "Invalidate authoring action",
                    $message: "At leat one field has to be displayed"
                });
                return;
            }
            var metadata = {};
            metadata[$bind] = {
                $isHidden: $isHidden
            };
            this.list.applyDesignMetaData({
                $item: {
                    $: metadata
                }
            }, true);
        }
        ($isHidden ? this.displayList : this.availableList).$$ul.find("[data-s-bind='" + $bind + "']").remove();
        this._addItem(($isHidden ? this.availableList : this.displayList), $bind, this.list.$fields[$bind]);
    },
    _addItem: function(itemList, $bind, $field){
        $field = $field || this.$fields[$bind];
        var li = document.createElement("li");
        li.className = "s-list-design-fields-li";
        li.setAttribute("data-s-bind", $bind);
        var $$li = $(li).appendTo(itemList.$$ul).text($field.$title);
        var btn = document.createElement("a");
        btn.className = itemList.$btnCss;
        $(btn).prependTo($$li);
    },
    _fillList: function(itemList){
        var self = this;
        itemList.$$tdBody = $("<div class='s-list-design-fields-td-body'/>").appendTo(itemList.$$td);
        itemList.$$ul = $("<ul class='s-list-design-fields-ul'/>").appendTo(itemList.$$tdBody);
        self.list.$item.$layout.$items.forEach(function($item){
            var $field = self.list.$fields[$item.$bind];
            if ($field && !$field.$isHidden) {
                if (itemList.$isHiddenItem == ($item.$isHidden || false)) {
                    self._addItem(itemList, $item.$bind, $field);
                }
            }
        });
        var btn = document.createElement("a");
        btn.className = "s-list-design-fields-select-all";
        $(btn).prependTo(itemList.$$th);
    },
    dispose: function(){
        delete this.list;
        if (this.$$item) {
            this.$$item.undelegate();
        }
        if (this.$$select) {
            this.$$select.unbind();
            this.$$unselect.unbind();
        }
    }
});
