// syracuse-x3/lib/convergence/client/CvgAction.js
"use strict";

var CvgIst = require('syracuse-x3/lib/convergence/types/CvgIst').CvgIst;

/**
 * Définition des actions Adonix
 *
 * <ul>
 * <li>** SRV__MINI 1024;
 * <li>
 * SRV_TIMEOUT..............1024..Action.serveur......1024..ACK_HGUP..0.Timeout
 * <li>SRV_CHAMP_SUIVANT........1025..Action.serveur......1024..ACK_ZSV1..1
 * <li>SRV_BOITE_SELECTION......1026..Action.serveur......1024..ACK_CHXL..2
 * <li>SRV_SORTIE_CHAMP_SPE.....1027..Action.serveur......1024..ACK_ZSV2..3
 * <li>SRV_VALIDE_ECRAN.........1028..Action.serveur......1024..ACK_VALP..4
 * <li>
 * SRV_ABANDON_SAISIE.......1029..Action.serveur......1024..ACK_ZPCD..5.aurait
 * .été.inversé.avec.SRV_CHAMP_PRECEDENT
 * <li>SRV_ABANDONLINE..........1030..Action.serveur......1024..ACK_FIN1..6
 * <li>SRV_ABANDON..............1031..Action.serveur......1024..ACK_FIN2..7
 * <li>SRV_DEBUGGER.............1040..Action.serveur......1024..ACK_DBGR..16
 * <li>SRV_CALCULATRICE.........1041..Action.serveur......1024..ACK_CALC..17
 * <li>SRV_GETFOCUS.............1044..Action.serveur......1024..ACK_CLIC..20
 * <li>SRV_COMMANDGRID..........1045..Action.serveur......1024..ACK_NAVI..21
 * <li>SRV_GETFOCUSCELL.........1050..Action.serveur......1024..ACK_MODC..26
 * <li>SRV_BRWCHANGELINE........1051..Action.serveur......1024..ACK_CLST..27
 * <li>SRV_BRWSELECTLINE........1052..Action.serveur......1024..ACK_SLST..28
 * <li>SRV_CHAMP_ONGLET.........1053..Action.serveur......1024..ACK_GONG..29
 * <li>SRV_CHAMP_PRECEDENT......1055..Action.serveur......1024..ACK_ZPC2..31..
 * aurait.été.inversé.avec.SRV_ABANDON_SAISIE
 * <li>SRV_BRWBEGINEDIT.........1056..Action.serveur......1024..ACK_MLST..32
 * <li>SRV_BRWENDEDIT...........1057..Action.serveur......1024..ACK_FLST..33
 * <li>SRV_CHAMP_INACTIF........1058..Action.serveur......1024..ACK_GRIS..34
 * <li>SRV_TREECHANGELINE.......1060..Action.serveur......1024..ACK_CABR..36
 * <li>SRV_TREESELECTLINE.......1061..Action.serveur......1024..ACK_SABR..37
 * <li>SRV_TREEEXPANDLINE.......1062..Action.serveur......1024..ACK_DABR..38
 * <li>SRV_TREEBEGINEDIT........1063..Action.serveur......1024..ACK_MABR..39
 * <li>SRV_TREEENDEDIT..........1064..Action.serveur......1024..ACK_FABR..40
 * <li>SRV_TREEDESELECTLINE.....1065..Action.serveur......1024..ACK_UABR..41
 * <li>SRV_PICKING..............1066..Action.serveur......1024..ACK_PICK..42
 * <li>SRV_BRWDESELECTLINE......1067..Action.serveur......1024..ACK_ULST..43
 * <li>SRV_FERME_PICKING........1068..Action.serveur......1024..ACK_FPCK..44
 * <li>SRV_BRWRAPIDESEL.........1069..Action.serveur......1024..ACK_SLRP..45
 * <li>SRV_CTL_FORMULE..........1070..Action.serveur......1024..ACK_CTRL..46
 * <li>SRV_ELEMENT_FORMULE......1071..Action.serveur......1024..ACK_TELT..47
 * <li>SRV_LIST_PAGE_UP.........1072..Action.serveur......1024..ACK_REWD..48
 * <li>SRV_LIST_PAGE_DOWN.......1073..Action.serveur......1024..ACK_FRWD..49
 * <li>SRV_OPEN_CTXMENU.........1074..Action.serveur......1024..ACK_LDMS..50
 * <li>
 * SRV_FIN_APPL.............1075..Action.serveur......1024..ACK_TERM..51..fin
 * .normale
 * <li>SRV_EXECFUNC.............1076..Action.serveur......1024..ACK_EXEF..52..
 * Execution.fonction.interactive
 * <li>SRV_CALLREMOTESPGM.......1077..Action.serveur......1024..ACK_CALL..53
 * <li>SRV_CALLREMOTEFUNC.......1078..Action.serveur......1024..ACK_FUNC..54
 * <li>SRV_ASKUI_RESPONSE.......1080..Action.serveur......1024..ACK_RESP. 56
 * <li>SRV_EXECI_RESPONSE.......1081..Action.serveur......1024..ACK_RESI..57
 * 15w_002 Fiche 34467 - Protocole: Support du noeud NDEXECI
 * <li>SRV_EXPXLSGRID.......1087..Action.serveur......1024..ACK_???...
 * <li>
 * SRV_DELETELINEGRID.......1089..Action.serveur......1024..ACK_SULI..65..A.:
 * .Cancel.Line
 * <li>
 * SRV_COPYLINEGRID.........1090..Action.serveur......1024..ACK_CPBF..66..B.:
 * .Bufferisation.Lignes
 * <li>
 * SRV_PASTELINEGRID........1091..Action.serveur......1024..ACK_RCBF..67..C.:
 * .Coller.du.Buffer
 * <li>
 * SRV_CUTLINEGRID..........1092..Action.serveur......1024..ACK_SUBK..68..D.:
 * .Cancel.a.Group.of.Line
 * <li>
 * SRV_INSERTLINEGRID.......1097..Action.serveur......1024..ACK_INLI..73..I.:
 * .Insert.Line
 * <li>
 * SRV_DELETELINESGRID......1107..Action.serveur......1024..ACK_DLBF..83...S.
 * :.Suppression
 * <li>SRV_EXECUTE..............1112..Action.serveur......1024..ACK_EXEC..88
 * <li>SRV_FICHIER_NOUVEAU......1115..Action.serveur......1024..ACK_NOUV..91
 * <li>SRV_FICHIER_ENREGISTRE...1116..Action.serveur......1024..ACK_ENRG..92
 * <li>SRV_FICHIER_CREATION.....1117..Action.serveur......1024..ACK_CREA..93
 * <li>SRV_FICHIER_SUPPRIME.....1118..Action.serveur......1024..ACK_EFFA..94
 * <li>SRV_SEL_SELECTION........1119..Action.serveur......1024..ACK_REQU..95
 * <li>SRV_FICHIER_PREMIER......1120..Action.serveur......1024..ACK_FDEB..96
 * <li>SRV_FICHIER_DERNIER......1121..Action.serveur......1024..ACK_FFIN..97
 * <li>SRV_FICHIER_SUIVANT......1122..Action.serveur......1024..ACK_FSUI..98
 * <li>SRV_FICHIER_PRECEDENT....1123..Action.serveur......1024..ACK_FPCD..99
 * <li>SRV_HELP................1036..Action.serveur......1024..ACK_AIDE..12
 * <li>SRV_CLIC_IMAGE...........1202..Action.serveur......1024..ACK_KIB...178
 * <li>SRV_CONNECTION...........1124..Action.serveur......1024..ACK_CNX...100 *
 * impémenté dans de serveur Adonix Java
 * <li>** SRV__MAXI 1279;
 * </ul>
 * <ul>
 * <li>SRV_SEL_SEL_MOT_CLE......2124..Action.superviseur..1024............1100
 * <li>SRV_SEL_CHANGE_MOT_CLE...2817..Action.superviseur..1024............1793
 * <li>SRV_FICHIER_PIECE_JOINT..2818..Action.superviseur..1024............1794
 * <li>SRV_FICHIER_COMMENTAIRE..2819..Action.superviseur..1024............1795
 * <li>SRV_FICHIER_PRINT........2820..Action.superviseur..1024............1796
 * <li>SRV_FICHIER_PRINT_LISTE..2821..Action.superviseur..1024............1797
 * <li>SRV_FICHIER_PROPRIETE....2822..Action.superviseur..1024............1798
 * <li>SRV_FICHIER_STATISTIQUE..2847..Action.superviseur..1024............1823
 * <li>SRV_FICHIER_WORKFLOW.....2848..Action.superviseur..1024............1824
 * <li>SRV_FICHIER_LIAISON......2849..Action.superviseur..1024............1825
 * </ul>
 *
 * @author Adonix Grenoble
 * @version 140_000
 */
exports.Acks = {
	// Sortie de l'Application
	ACK_HGUP: 0,
	// Zone Suivante (1)
	ACK_ZSV1: 1,
	// Demande de Selection
	ACK_CHXL: 2,
	// Zone Suivante (2)
	ACK_ZSV2: 3,
	// Validation de la Page
	ACK_VALP: 4,
	// Zone Precedente
	ACK_ZPCD: 5,
	// Abandon de la Page (1)
	ACK_FIN1: 6,
	// Abandon de la Page (2)
	ACK_FIN2: 7,
	// Touche Fleche a Gauche
	ACK_KFLG: 8,
	// Touche Fleche a Droite
	ACK_KFLD: 9,
	// Touche Delete
	ACK_KDEL: 10,
	// Programmation Touches
	ACK_PFNC: 11,
	// Aide de l'Application
	ACK_AIDE: 12,
	// Touche Cancel
	ACK_KCAN: 13,
	// Touche Insertion
	ACK_KINS: 14,
	// Hardcopy
	ACK_HCPY: 15,
	// Debugger
	ACK_DBGR: 16,
	// Calculatrice
	ACK_CALC: 17,
	// Touche Espace
	ACK_KESP: 18,
	// Touche Tabulation
	ACK_KTAB: 19,
	// Passage a une Zone Donnee
	ACK_CLIC: 20,
	// Passage en Mode Navigation
	ACK_NAVI: 21,
	// Passage Ligne Suivante
	ACK_LSUI: 22,
	// Validation de la Ligne
	ACK_VALL: 23,
	// Passage Ligne Precedente
	ACK_LPCD: 24,
	// Abandon de la Ligne
	ACK_FINL: 25,
	// Modification Cellule
	ACK_MODC: 26,
	// Changement de Ligne Liste
	ACK_CLST: 27,
	// Selection d'un Ligne Liste
	ACK_SLST: 28,
	// Entree dans l'Onglet
	ACK_GONG: 29,
	// Chargement Combo Box
	ACK_LDLB: 30,
	// Zone Precedente Valide
	ACK_ZPC2: 31,
	// Modification de Ligne Liste
	ACK_MLST: 32,
	// Fin de Modif Ligne Liste
	ACK_FLST: 33,
	// L'Instance est Descativee
	ACK_GRIS: 34,
	// Clic qui Valide
	ACK_CLIV: 35,
	// Chgment de Ligne d'Arbre
	ACK_CABR: 36,
	// Selection Ligne d'Arbre
	ACK_SABR: 37,
	// Developpement ligne d'arbre
	ACK_DABR: 38,
	// Modification de Ligne Arbre
	ACK_MABR: 39,
	// Fin de Modif Ligne Arbre
	ACK_FABR: 40,
	// Deselection Ligne d'Arbre
	ACK_UABR: 41,
	// Action de Picking
	ACK_PICK: 42,
	// Deselection Ligne de Liste
	ACK_ULST: 43,
	// Fermeture Fenetre Picking
	ACK_FPCK: 44,
	// Selection Rapide de Liste
	ACK_SLRP: 45,
	// Controle Formule
	ACK_CTRL: 46,
	// Selection Rapide de Liste
	ACK_TELT: 47,
	// Revenir en Arriere
	ACK_REWD: 48,
	// Avancer
	ACK_FRWD: 49,
	// demande items menu click droit
	ACK_LDMS: 50,
	// fin normale
	ACK_TERM: 51,
	// Demande execution fonction interactive depuis le menu
	ACK_MENU: 52,
	// Demande exécution sous programme catalogué
	ACK_CALL: 53,
	// Demande exécution function cataloguée
	ACK_FUNC: 54,
	// @version 142 - Réponse à un AskUi
	ACK_RESP: 56,
	// 15w_002 Fiche 34467 - Protocole: Support du noeud NDEXECI
	// Réponse à un NDEXECI
	ACK_RESI: 57,
	// A : Cancel Line
	ACK_SULI: 65,
	// B : Bufferisation Ligne s
	ACK_CPBF: 66,
	// C : Coller du Buffer
	ACK_RCBF: 67,
	// D : Cancel a Group of L ine
	ACK_SUBK: 68,
	// G : Navigation
	ACK_NAV2: 71,
	// I : Insert Line
	ACK_INLI: 73,
	// K : Modif d'une Ligne
	ACK_MOD1: 75,
	// M : Modif des Lignes
	ACK_MODI: 77,
	// R : Resume
	ACK_REPR: 82,
	// S : Supprssion
	ACK_DLBF: 83,
	// X : Execution Traitemen t
	ACK_EXEC: 88,
	// Option Nouveau
	ACK_NOUV: 91,
	// Option Enregistrer
	ACK_ENRG: 92,
	// Option Creation
	ACK_CREA: 93,
	// Option Effacement
	ACK_EFFA: 94,
	// Option Requete
	ACK_REQU: 95,
	// Option Premier
	ACK_FDEB: 96,
	// Option Dernier
	ACK_FFIN: 97,
	// Option Suivant
	ACK_FSUI: 98,
	// Option Precedent
	ACK_FPCD: 99,
	// 16j_000 Connexion
	ACK_CNX: 100,
	//
	ACK_KF01: 129,
	//
	ACK_KF02: 130,
	//
	ACK_KF03: 131,
	//
	ACK_KF04: 132,
	//
	ACK_KF05: 133,
	//
	ACK_KF06: 134,
	//
	ACK_KF07: 135,
	//
	ACK_KF08: 136,
	//
	ACK_KF09: 137,
	//
	ACK_KF10: 138,
	//
	ACK_KF11: 139,
	//
	ACK_KF12: 140,
	//
	ACK_KF13: 141,
	//
	ACK_KF14: 142,
	//
	ACK_KF15: 143,
	//
	ACK_KF16: 144,
	//
	ACK_KF17: 145,
	//
	ACK_KF18: 146,
	//
	ACK_KF19: 147,
	//
	ACK_KF20: 148,
	// Clic sur une image
	ACK_KIB: 178,
	//
	ACK_STIN: 201,
	//
	ACK_STOV: 202,
	//
	ACK_CONT: 203,
	// On Recommence
	ACK_BCLE: 255,
};

var RAW_ACTIONS = [
	// Action serveur 1024 + ACK_FIN2 (7)
	['SRV_ABANDON', 1031, ''],
	// Action serveur 1024 + ACK_ZPCD (5) : aurait été inversé avec SRV_CHAMP_PRECEDENT
	['SRV_ABANDON_SAISIE', 1029, ''],
	// Action serveur 1024 + ACK_FIN1 (6)
	['SRV_ABANDONLINE', 1030, ''],
	// Action serveur 1024 ACK_RESP (56)
	['SRV_ASKUI_RESPONSE', 1080, ''],
	// Action serveur 1024 ACK_RESI (57) 15w_002 Fiche 34467 - Protocole: Support du noeud
	['SRV_EXECI_RESPONSE', 1081, 'ISSUBUxVxYxW'],
	// Action serveur 1024 + ACK_CHXL (2)
	['SRV_BOITE_SELECTION', 1026, ''],
	// Action serveur 1024 + ACK_MLST (32)
	['SRV_BRWBEGINEDIT', 1056, 'I'],
	// Action serveur 1024 + ACK_CLST (27)
	['SRV_BRWCHANGELINE', 1051, 'IW'],
	// Action serveur 1024 + ACK_ULST (43)
	['SRV_BRWDESELECTLINE', 1067, 'IW'],
	// Action serveur 1024 + ACK_FLST (33)
	['SRV_BRWENDEDIT', 1057, 'I'],
	// Action serveur 1024 + ACK_SLRP (45)
	['SRV_BRWRAPIDESEL', 1069, 'IW'],
	// Action serveur 1024 + ACK_SLST (28)
	['SRV_BRWSELECTLINE', 1052, 'IW'],
	// Action serveur 1024 + ACK_CALC (17)
	['SRV_CALCULATRICE', 1041, ''], // SPECIAL!
	// Action serveur 1024 + ACK_FUNC (54)... ie SRV_CALLREMOTESPGM
	['SRV_CALLREMOTEFUNC', 1078, 'IULxTxW'],
	// Action serveur 1024 + ACK_CALL (53)
	['SRV_CALLREMOTESPGM', 1077, 'IULxTxW'], // x mean the next descriptor describe an extends
	// Action serveur 1024 + ACK_GRIS (34)
	['SRV_CHAMP_INACTIF', 1058, ''],
	// Action serveur 1024 + ACK_GONG (29)
	['SRV_CHAMP_ONGLET', 1053, ''],
	// Action serveur 1024 + ACK_ZPC2 (31) : aurait été inversé avec SRV_ABANDON_SAISIE
	['SRV_CHAMP_PRECEDENT', 1055, ''],
	// Action serveur 1024 + ACK_ZSV1 (1)
	['SRV_CHAMP_SUIVANT', 1025, ''],
	// Action serveur 1024 + ACK_KIB (178)
	['SRV_CLIC_IMAGE', 1202, 'I'],
	// Action serveur 1024 + ACK_NAVI (21)
	['SRV_COMMANDGRID', 1045, 'I'],
	// Action serveur 1024 + ACK_CPBF (66) : B : Bufferisation Lignes
	['SRV_COPYLINEGRID', 1090, 'ISS'],
	// Action serveur 1024 + ACK_CTRL (46)... Warning one parameter ==> Key of the item to load (new arg serializer)
	['SRV_CTL_FORMULE', 1070, 'U'],
	// Action serveur 1024 + ACK_SUBK (68) : D : Cancel a Group of Line
	['SRV_CUTLINEGRID', 1092, 'ISS'],
	// Action serveur 1024 + ACK_DBGR (16)
	['SRV_DEBUGGER', 1040, ''],
	// Action serveur 1024 + ACK_SULI (65) : A : Cancel Line
	['SRV_DELETELINEGRID', 1089, 'I'],
	// Action serveur 1024 + ACK_DLBF (83) : S : Suppression
	['SRV_DELETELINESGRID', 1107, 'ISS'],
	// Action serveur 1024 + ACK_TELT (47)... Warning one parameter ==> Key of the item to load (new arg serializer)
	['SRV_ELEMENT_FORMULE', 1071, 'U'],
	// Action serveur 1024 + ACK_EXEF (52) : Execution fonction interactive... normaly never sent in fusion context.
	['SRV_EXECFUNC', 1076, '?'],
	// Action serveur 1024 + ACK_EXEC (88)... SPECIAL : L4G process submission
	['SRV_EXECUTE', 1112, ''],
	// Action serveur 1024 + ACK_FPCK (44)
	['SRV_FERME_PICKING', 1068, 'I'],
	// Action serveur 1024 + ACK_CREA (93)
	['SRV_FICHIER_CREATION', 1117, ''],
	// Action serveur 1024 + ACK_FFIN (97)
	['SRV_FICHIER_DERNIER', 1121, ''],
	// Action serveur 1024 + ACK_ENRG (92)
	['SRV_FICHIER_ENREGISTRE', 1116, ''],
	// Action serveur 1024 + ACK_NOUV (91)
	['SRV_FICHIER_NOUVEAU', 1115, ''],
	// Action serveur 1024 + ACK_FPCD (99)
	['SRV_FICHIER_PRECEDENT', 1123, ''],
	// Action serveur 1024 + ACK_FDEB (96)
	['SRV_FICHIER_PREMIER', 1120, ''],
	// Action serveur 1024 + ACK_FSUI (98)
	['SRV_FICHIER_SUIVANT', 1122, ''],
	// Action serveur 1024 + ACK_EFFA (94)
	['SRV_FICHIER_SUPPRIME', 1118, ''],
	// Action serveur 1024 + ACK_TERM (51) : fin normale
	['SRV_FIN_APPL', 1075, ''],
	// Action serveur 1024 + ACK_CLIC (20)
	['SRV_GETFOCUS', 1044, 'I'],
	// Action serveur 1024 + ACK_MODC (26)
	['SRV_GETFOCUSCELL', 1050, 'I'],
	// Action superviseur 1024 + ACK_AIDE (12)... probably obsolete
	['SRV_HELP', 1036, ''],
	// Action serveur 1024 + ACK_INLI (73) : I : Insert Line
	['SRV_INSERTLINEGRID', 1097, 'IS'],
	// Action serveur 1024 + ACK_FRWD (49)
	['SRV_LIST_PAGE_DOWN', 1073, 'IW'], // ist + key
	// Action serveur 1024 + ACK_REWD (48)
	['SRV_LIST_PAGE_UP', 1072, 'IW'], // ist + optional key
	// Action serveur 1024 + ACK_LDMS (50)
	['SRV_OPEN_CTXMENU', 1074, 'I'],
	// Action serveur 1024 + ACK_RCBF (67) : C : Coller du Buffer
	['SRV_PASTELINEGRID', 1091, 'IS'],
	// Action serveur 1024 + ACK_PICK (42)
	['SRV_PICKING', 1066, 'I'],
	// Action serveur 1024 + ACK_REQU (95)... probably obsolete
	['SRV_SEL_SELECTION', 1119, ''],
	// Action serveur 1024 + ACK_ZSV2 (3)... probably obsolete
	['SRV_SORTIE_CHAMP_SPE', 1027, ''],
	// Action serveur 1024 + ACK_HGUP (0) : Timeout
	['SRV_TIMEOUT', 1024, ''],
	// Action serveur 1024 + ACK_MABR (39)
	['SRV_TREEBEGINEDIT', 1063, ''],
	// Action serveur 1024 + ACK_CABR (36)
	['SRV_TREECHANGELINE', 1060, 'IW'],
	// Action serveur 1024 + ACK_UABR (41)
	['SRV_TREEDESELECTLINE', 1065, 'IW'],
	// Action serveur 1024 + ACK_FABR (40)
	['SRV_TREEENDEDIT', 1064, ''],
	// Action serveur 1024 + ACK_DABR (38)
	['SRV_TREEEXPANDLINE', 1062, 'IW'],
	// Action serveur 1024 + ACK_SABR (37)
	['SRV_TREESELECTLINE', 1061, 'IW'],
	// Action serveur 1024 + ACK_VALP (4)
	['SRV_VALIDE_ECRAN', 1028, ''],
	// Action serveur 1024 + ACK_KF01 (129)
	['SRV_TUNNEL', 1153, ''],
	// Action serveur 1024 + (163)
	['SRV_EXPXLSGRID', 1187, '']
];

exports.Actions = RAW_ACTIONS.reduce(function(actions, desc) {
	actions[desc[0]] = desc[1];
	return actions;
}, {});

exports.Descriptors = RAW_ACTIONS.reduce(function(actions, desc) {
	actions[desc[1]] = {
		name: desc[0],
		id: desc[1],
		params: desc[2],
	};
	return actions;
}, {});

var processNextAction = function(actNum, reqTarg, respTarg, checkNumLine) {
	// Compare win/xid/nl
	return respTarg !== null && respTarg.win === reqTarg.win && //
	respTarg.xid === reqTarg.xid && (!checkNumLine || (respTarg.nl === reqTarg.nl));
};

exports.SuperActions = [{
	id: "not used (0)",
	code: 0,
	actions: [],
	processNextAction: processNextAction,
}, {
	id: "not used (769)",
	code: 769,
	actions: [],
	processNextAction: processNextAction,
}, {
	id: "not used (770)",
	code: 770,
	actions: [],
	processNextAction: processNextAction,
}, {
	id: "SUP_FIELD_DRILL",
	code: 771,
	actions: [{
		name: 'SRV_GETFOCUS',
		setTarget: true,
	}, {
		name: 'SRV_TUNNEL',
		setTarget: true,
	}],
	processNextAction: processNextAction,
}, {
	id: "SUP_FIELD_ICOCLICK",
	code: 772,
	actions: [{
		name: 'SRV_GETFOCUS',
		setTarget: true,
	}, {
		name: 'SRV_CLIC_IMAGE',
		setTarget: true,
	}],
	processNextAction: processNextAction,
}, {
	id: "SUP_FIELD_CTX",
	code: 773,
	actions: [{
		name: 'SRV_GETFOCUS',
		setTarget: true,
	}, {
		name: 'SRV_OPEN_CTXMENU',
		setTarget: true,
	}],
	processNextAction: processNextAction,
}, {
	id: "SUP_CELL_EDIT",
	code: 774,
	actions: [{
		name: 'SRV_COMMANDGRID',
		setTarget: true,
		checkNumLine: true,
	}, {
		name: 'SRV_GETFOCUSCELL',
		setTarget: true,
		checkNumLine: true,
	}],
	processNextAction: processNextAction,
}, {
	id: "SUP_CELL_ICOCLICK",
	code: 775,
	actions: [{
		name: 'SRV_COMMANDGRID',
		setTarget: true,
		checkNumLine: true,
	}, {
		name: 'SRV_GETFOCUSCELL',
		setTarget: true,
		checkNumLine: true,
	}, {
		name: 'SRV_CLIC_IMAGE',
		setTarget: true,
		checkNumLine: true,
	}, {
		name: 'SRV_COMMANDGRID',
		setTarget: true,
		checkNumLine: true,
	}],
	processNextAction: processNextAction,
}, {
	id: "SUP_CELL_ICOCLICKFOCUS",
	code: 776,
	actions: [{
		name: 'SRV_GETFOCUSCELL',
		setTarget: true,
		checkNumLine: true,
	}, {
		name: 'SRV_CLIC_IMAGE',
		setTarget: true,
		checkNumLine: true,
	}, {
		name: 'SRV_COMMANDGRID',
		setTarget: true,
		checkNumLine: true,
	}],
	processNextAction: processNextAction,
}, {
	id: "SUP_CELL_CHECK", //unused
	code: 777,
	deprecated: true,
}, {
	id: "SUP_CELL_CHECKFOCUS", //unused
	code: 778,
	deprecated: true,
}, {
	id: "SUP_LIST_PREV",
	code: 779,
	actions: [{
		name: 'SRV_GETFOCUS',
		setTarget: true,
	}, {
		name: 'SRV_LIST_PAGE_UP',
		setTarget: true,
	}],
	processNextAction: processNextAction,
}, {
	id: "SUP_LIST_NEXT",
	code: 780,
	actions: [{
		name: 'SRV_GETFOCUS',
		setTarget: true,
	}, {
		name: 'SRV_LIST_PAGE_DOWN',
		setTarget: true,
	}],
	processNextAction: processNextAction,
}, {
	id: "SUP_LIST_SELLINE",
	code: 781,
	actions: [{
		name: 'SRV_GETFOCUS',
		setTarget: true,
	}, {
		name: 'SRV_BRWCHANGELINE',
		setTarget: true,
	}],
	processNextAction: processNextAction,
}, {
	id: "SUP_LIST_FILTER",
	code: 782,
	actions: [{
		name: 'SRV_BRWRAPIDESEL',
		setTarget: true,
	}, {
		name: 'SRV_GETFOCUS',
		setTarget: true,
	}],
	processNextAction: function(actNum, reqTarg, respTarg, checkNumLine) {
		return true;
	},
}, {
	id: "SUP_WIN_CLOSESTACK",
	code: 783,
	actions: [],
}, {
	id: "SUP_LIST_SELLINEEX",
	code: 784,
	actions: [{
		name: 'SRV_GETFOCUS',
		setTarget: true,
	}, {
		name: 'SRV_BRWSELECTLINE',
		setTarget: true,
	}],
	processNextAction: processNextAction,
}, {
	id: "SUP_FIELD_SELECT",
	code: 785,
	actions: [{
		name: 'SRV_GETFOCUS',
		setTarget: true,
	}, {
		name: 'SRV_BOITE_SELECTION',
	}],
	processNextAction: processNextAction,
}, {
	id: "SUP_LINE_DELETE",
	code: 786,
	actions: [{
		name: 'SRV_COMMANDGRID',
		setTarget: true,
		checkNumLine: true,

	}, {
		name: 'SRV_DELETELINEGRID',
		setTarget: true,
	}],
	processNextAction: function(actNum, reqTarg, respTarg, checkNumLine) {
		var check = processNextAction(actNum, reqTarg, respTarg, checkNumLine);
		if (actNum === 1045 && !check) {
			return reqTarg.win === respTarg.win && reqTarg.blocAlphaId === respTarg.blocAlphaId && reqTarg.nl === respTarg.nl;
		}
		return check;
	},
}, {
	id: "SUP_LINE_INSERT",
	code: 787,
	actions: [{
		name: 'SRV_COMMANDGRID',
		setTarget: true,
		checkNumLine: true,
	}, {
		name: 'SRV_INSERTLINEGRID',
		setTarget: true,
	}],
	processNextAction: function(actNum, reqTarg, respTarg, checkNumLine) {
		var check = processNextAction(actNum, reqTarg, respTarg, checkNumLine);
		if (actNum === 1045 && !check) {
			return reqTarg.win === respTarg.win && reqTarg.blocAlphaId === respTarg.blocAlphaId && reqTarg.nl === respTarg.nl;
		}
		return check;
	},
}, {
	id: "SUP_LINES_DELETE",
	code: 788,
	actions: [{
		name: 'SRV_COMMANDGRID',
		setTarget: true,
		checkNumLine: true,
	}, {
		name: 'SRV_DELETELINESGRID',
		setTarget: true,
	}],
	processNextAction: function(actNum, reqTarg, respTarg, checkNumLine) {
		var check = processNextAction(actNum, reqTarg, respTarg, checkNumLine);
		if (actNum === 1045 && !check) {
			return reqTarg.win === respTarg.win && reqTarg.blocAlphaId === respTarg.blocAlphaId && reqTarg.nl === respTarg.nl;
		}
		return check;
	},
}, {
	id: "SUP_CELL_FOCUS_DRILL",
	code: 789,
	actions: [{
		name: 'SRV_COMMANDGRID',
		setTarget: true,
		checkNumLine: true,
	}, {
		name: 'SRV_TUNNEL',
		setTarget: true,
	}],
	processNextAction: function(actNum, reqTarg, respTarg, checkNumLine) {
		var check = processNextAction(actNum, reqTarg, respTarg, checkNumLine);
		if (check) {
			return reqTarg.nl === respTarg.nl;
		}
		return check;
	},
}, {
	id: "SUP_CELL_EDIT_DRILL",
	code: 790,
	actions: [{
		name: 'SRV_GETFOCUSCELL',
		setTarget: true,
		checkNumLine: true,
	}, {
		name: 'SRV_TUNNEL',
		setTarget: true,
	}],
	processNextAction: function(actNum, reqTarg, respTarg, checkNumLine) {
		var check = processNextAction(actNum, reqTarg, respTarg, checkNumLine);
		if (check) {
			return reqTarg.nl === respTarg.nl;
		}
		return check;
	},
}, {
	id: "SUP_CELL_FOCUS_EDIT_DRILL",
	code: 791,
	actions: [{
		name: 'SRV_COMMANDGRID',
		setTarget: true,
		checkNumLine: true,
	}, {
		name: 'SRV_GETFOCUSCELL',
		setTarget: true,
		checkNumLine: true,
	}, {
		name: 'SRV_TUNNEL',
		setTarget: true,
	}],
	processNextAction: function(actNum, reqTarg, respTarg, checkNumLine) {
		var check = processNextAction(actNum, reqTarg, respTarg, checkNumLine);
		if (check) {
			return reqTarg.nl === respTarg.nl;
		}
		return check;
	},
}, {
	id: "SUP_CELL_FOCUS_SELECT",
	code: 792,
	actions: [{
		name: 'SRV_COMMANDGRID',
		setTarget: true,
		checkNumLine: true,
	}, {
		name: 'SRV_BOITE_SELECTION',
		setTarget: true,
	}],
	processNextAction: function(actNum, reqTarg, respTarg, checkNumLine) {
		var check = processNextAction(actNum, reqTarg, respTarg, checkNumLine);
		if (check) {
			return reqTarg.nl === respTarg.nl;
		}
		return check;
	},
}, {
	id: "SUP_CELL_EDIT_SELECT",
	code: 793,
	actions: [{
		name: 'SRV_GETFOCUSCELL',
		setTarget: true,
		checkNumLine: true,
	}, {
		name: 'SRV_BOITE_SELECTION',
		setTarget: true,
	}],
	processNextAction: function(actNum, reqTarg, respTarg, checkNumLine) {
		var check = processNextAction(actNum, reqTarg, respTarg, checkNumLine);
		if (check) {
			return reqTarg.nl === respTarg.nl;
		}
		return check;
	},
}, {
	id: "SUP_CELL_FOCUS_EDIT_SELECT",
	code: 794,
	actions: [{
		name: 'SRV_COMMANDGRID',
		setTarget: true,
		checkNumLine: true,
	}, {
		name: 'SRV_GETFOCUSCELL',
		setTarget: true,
		checkNumLine: true,
	}, {
		name: 'SRV_BOITE_SELECTION',
		setTarget: true,
	}],
	processNextAction: function(actNum, reqTarg, respTarg, checkNumLine) {
		var check = processNextAction(actNum, reqTarg, respTarg, checkNumLine);
		if (check) {
			return reqTarg.nl === respTarg.nl;
		}
		return check;
	},
}, {
	id: "SUP_CELL_FOCUS_CTX",
	code: 795,
	actions: [{
		name: 'SRV_COMMANDGRID',
		setTarget: true,
		checkNumLine: true,
	}, {
		name: 'SRV_OPEN_CTXMENU',
		setTarget: true,
	}],
	processNextAction: function(actNum, reqTarg, respTarg, checkNumLine) {
		var check = processNextAction(actNum, reqTarg, respTarg, checkNumLine);
		if (check) {
			return reqTarg.nl === respTarg.nl;
		} else {
			return reqTarg.win === respTarg.win && reqTarg.blocAlphaId === respTarg.blocAlphaId && reqTarg.nl === respTarg.nl;
		}
	},
}, {
	id: "SUP_CELL_EDIT_CTX",
	code: 796,
	actions: [{
		name: 'SRV_GETFOCUSCELL',
		setTarget: true,
		checkNumLine: true,
	}, {
		name: 'SRV_OPEN_CTXMENU',
		setTarget: true,
	}],
	processNextAction: function(actNum, reqTarg, respTarg, checkNumLine) {
		var check = processNextAction(actNum, reqTarg, respTarg, checkNumLine);
		if (check) {
			return reqTarg.nl === respTarg.nl;
		}
		return check;
	},
}, {
	id: "SUP_CELL_FOCUS_EDIT_CTX",
	code: 797,
	actions: [{
		name: 'SRV_COMMANDGRID',
		setTarget: true,
		checkNumLine: true,
	}, {
		name: 'SRV_GETFOCUSCELL',
		setTarget: true,
		checkNumLine: true,
	}, {
		name: 'SRV_OPEN_CTXMENU',
		setTarget: true,
	}],
	processNextAction: function(actNum, reqTarg, respTarg, checkNumLine) {
		var check = processNextAction(actNum, reqTarg, respTarg, checkNumLine);
		if (check) {
			return reqTarg.nl === respTarg.nl;
		}
		return check;
	},
}, {
	id: "SUP_LIST_PICKING_FOCUS_SELECT", //set focus and select picking one line
	code: 798,
	actions: [{
		name: 'SRV_GETFOCUS',
		setTarget: true,
	}, {
		name: 'SRV_PICKING',
		setTarget: true,
		propagate: 'SRV_TREESELECTLINE',
	}],
	processNextAction: function(actNum, reqTarg, respTarg, checkNumLine) {
		var check = processNextAction(actNum, reqTarg, respTarg, checkNumLine);
		if (actNum === 1066) {
			return true;
		}
		return check;
	},
}, {
	id: "SUP_LIST_PICKING_SELECT", //select picking one line
	code: 799,
	actions: [{
		name: 'SRV_PICKING',
		setTarget: true,
		propagate: 'SRV_TREESELECTLINE',
	}],
	processNextAction: function(actNum, reqTarg, respTarg, checkNumLine) {
		return true;
	},
}, {
	id: "SUP_LIST_PICKING_FOCUS_UNSELECT", //set focus and unselect picking one line
	code: 800,
	actions: [{
		name: 'SRV_GETFOCUS',
		setTarget: true,
	}, {
		name: 'SRV_PICKING',
		setTarget: true,
		propagate: 'SRV_TREEDESELECTLINE',
	}],
	processNextAction: function(actNum, reqTarg, respTarg, checkNumLine) {
		var check = processNextAction(actNum, reqTarg, respTarg, checkNumLine);
		if (actNum === 1066) {
			return true;
		}
		return check;
	},
}, {
	id: "SUP_LIST_PICKING_UNSELECT", //unselect picking one line
	code: 801,
	actions: [{
		name: 'SRV_PICKING',
		setTarget: true,
		propagate: 'SRV_TREEDESELECTLINE',
	}],
	processNextAction: function(actNum, reqTarg, respTarg, checkNumLine) {
		return true;
	},
}, {
	id: "SUP_CELL_EDIT_TAB", //set edit and tab
	code: 802,
	actions: [{
		name: 'SRV_GETFOCUSCELL',
		setTarget: true,
		checkNumLine: true,
	}, {
		name: 'SRV_CHAMP_SUIVANT',
		setTarget: true,
	}],
	processNextAction: function(actNum, reqTarg, respTarg, checkNumLine) {
		var check = processNextAction(actNum, reqTarg, respTarg, checkNumLine);
		if (check) {
			return reqTarg.nl === respTarg.nl;
		}
		return check;
	},
}, {
	id: "SUP_LIST_TREESELECTLINE", //select leaf recursive hierarchy 
	code: 803,
	actions: [{
		name: 'SRV_GETFOCUS',
		setTarget: true,
	}, {
		name: 'SRV_TREESELECTLINE',
		setTarget: true,
	}],
	processNextAction: function(actNum, reqTarg, respTarg, checkNumLine) {
		return true;
	},
}, {
	id: "SUP_LIST_TREECHANGELINE", //select leaf hierarchy 
	code: 804,
	actions: [{
		name: 'SRV_GETFOCUS',
		setTarget: true,
	}, {
		name: 'SRV_TREECHANGELINE',
		setTarget: true,
	}],
	processNextAction: function(actNum, reqTarg, respTarg, checkNumLine) {
		return true;
	},
}, {
	id: "SUP_LIST_TREEEXPANDLINE", //expend line hierarchy 
	code: 805,
	actions: [{
		name: 'SRV_GETFOCUS',
		setTarget: true,
	}, {
		name: 'SRV_TREEEXPANDLINE',
		setTarget: true,
	}],
	processNextAction: function(actNum, reqTarg, respTarg, checkNumLine) {
		return true;
	},
}, {
	id: "SUP_EXPXLSGRID",
	code: 806,
	actions: [{
		name: 'SRV_COMMANDGRID',
		setTarget: true,
		checkNumLine: true,
	}, {
		name: 'SRV_EXPXLSGRID',
		setTarget: true,
	}],
	processNextAction: function(actNum, reqTarg, respTarg, checkNumLine) {
		return true;
	},
}, {
	id: "SUP_LIST_ABANDON",
	code: 807,
	actions: [{
		name: 'SRV_ABANDONLINE',
		setTarget: true,
		checkNumLine: false

	}, {
		name: 'SRV_CHAMP_SUIVANT',
		setTarget: true,
	}],
	processNextAction: function(actNum, reqTarg, respTarg, checkNumLine) {
		return true;
	},
}, {
	id: "SUP_FIELD_SRVCALL",
	code: 808,
	actions: [{
		name: 'SRV_GETFOCUS',
		setTarget: true
	}, {
		name: 'SRV_CALLREMOTESPGM',
		setTarget: true
	}],
	processNextAction: processNextAction
}, {
	id: "SUP_CELL_FOCUS_SRVCALL",
	code: 809,
	actions: [{
		name: 'SRV_COMMANDGRID',
		setTarget: true,
		checkNumLine: true
	}, {
		name: 'SRV_CALLREMOTESPGM',
		setTarget: true
	}],
	processNextAction: function(actNum, reqTarg, respTarg, checkNumLine) {
		var check = false,
			istReq = CvgIst.fromAction(respTarg),
			istResp = CvgIst.fromAction(respTarg);
		if (reqTarg.nl > 0) {
			check = processNextAction(actNum, reqTarg, respTarg, checkNumLine);
		} else {
			check = istReq.win === istResp.win && istReq.screen === istResp.screen && istReq.bloc === istResp.bloc;
		}
		return check;
	}
}, {
	id: "SUP_FIELD_SRVFUNC",
	code: 810,
	actions: [{
		name: 'SRV_GETFOCUS',
		setTarget: true
	}, {
		name: 'SRV_CALLREMOTEFUNC',
		setTarget: true
	}],
	processNextAction: processNextAction
}, {
	id: "SUP_CELL_FOCUS_SRVFUNC",
	code: 811,
	actions: [{
		name: 'SRV_COMMANDGRID',
		setTarget: true,
		checkNumLine: true
	}, {
		name: 'SRV_CALLREMOTEFUNC',
		setTarget: true
	}],
	processNextAction: function(actNum, reqTarg, respTarg, checkNumLine) {
		var check = false,
			istReq = CvgIst.fromAction(respTarg),
			istResp = CvgIst.fromAction(respTarg);
		if (reqTarg.nl > 0) {
			check = processNextAction(actNum, reqTarg, respTarg, checkNumLine);
		} else {
			check = istReq.win === istReq.win && istReq.screen === istResp.screen && istReq.bloc === istResp.bloc;
		}
		return check;
	}
}, {
	id: "SUP_WINS_CLOSE", //select leaf recursive hierarchy 
	code: 812,
	actions: [{
		isLoop: true,
		name: 'SRV_ABANDON',
		setTarget: true,
	}],
	processNextAction: function(actNum, reqTarg, respTarg, checkNumLine) {
		return true;
	},
}];

exports.SpecialActions = [{
	id: "not used (0)",
	code: 512,
	proxy: null,
}, {
	id: "PRINTER_SEL_OPTS",
	code: 513,
	proxy: "srvImp"
}, {
	id: "PRINTER_SEL_CANCEL",
	code: 514,
	proxy: "srvImp"
}, {
	id: "PRINTER_SEL_OK",
	code: 515,
	proxy: "srvImp"
}, {
	id: "PRINTER_OPTS_CANCEL",
	code: 516,
	proxy: "srvImp"
}, {
	id: "PRINTER_OPTS_OK",
	code: 517,
	proxy: "srvImp"
}, ];
exports.CONNECT_ACTION = 1;
