// syracuse-x3/lib/convergence/client/FusionClientInfo.js
"use strict";
// obsolete stuff - will go away
var helpers = require('syracuse-core/lib/helpers');

var FusionClientInfo = exports.FusionClientInfo = helpers.defineClass(function(msgCode, msgDef, errCode) {
	// Code du message utilisateur - pour traduction dans le client via un menu local
	this.msgCode = msgCode;
	// Code affiche par defaut si pas de traduction
	this.msgDef = msgDef;
	// Code erreur qui determine le comportement du client a adopter
	this.errCode = errCode !== undefined ? errCode : FusionErrCode.KILL;
}, null, {
	apendToJSon: function(o, errMsg) {
		if (o == null) {
			o = {};
		}
		o.msgCode = 500 + this.msgCode;
		o.msgDefault = this.msgDef;
		o.errCode = this.errCode.name();
		// Precision sur la cause de l erreur - Non exploite par le client mais interessant de visualiser dans les traces des requetes http
		o.errMsg = errMsg == null ? "" : errMsg;
		return o;
	},
});

var FusionErrCode = {
	KILL: 0,
	BUSY: 1
};

FusionClientInfo.NOCATCH = new FusionClientInfo(0, "Unexpected uncatched exception");
FusionClientInfo.JAVA = new FusionClientInfo(1, "Unexpected Java Exception");
FusionClientInfo.CREATECNX = new FusionClientInfo(2, "Create Syracuse connection  failed");
FusionClientInfo.CREATESESS = new FusionClientInfo(3, "Create X3 session failed");
FusionClientInfo.DELETECNX = new FusionClientInfo(4, "Delete Syracuse connection  failed");
FusionClientInfo.SESSNOTFOUND = new FusionClientInfo(5, "User session not found");
FusionClientInfo.SESSBUSY = new FusionClientInfo(6, "Can't process request - X3 Client is busy", FusionErrCode.BUSY);
FusionClientInfo.BADURL = new FusionClientInfo(7, "Bad url");
FusionClientInfo.REQHDL = new FusionClientInfo(8, "Error processing client request");
FusionClientInfo.X3COM = new FusionClientInfo(9, "X3 communication exception");
FusionClientInfo.CONFIG = new FusionClientInfo(10, "Java Bridge Server configuration exception");
