// hdb/lib-cov/util/bignum.js
/* automatically generated by JSCoverage - do not edit */
if (typeof _$jscoverage === 'undefined') _$jscoverage = {};
if (! _$jscoverage['util/bignum.js']) {
  _$jscoverage['util/bignum.js'] = [];
  _$jscoverage['util/bignum.js'][14] = 0;
  _$jscoverage['util/bignum.js'][17] = 0;
  _$jscoverage['util/bignum.js'][19] = 0;
  _$jscoverage['util/bignum.js'][21] = 0;
  _$jscoverage['util/bignum.js'][23] = 0;
  _$jscoverage['util/bignum.js'][24] = 0;
  _$jscoverage['util/bignum.js'][25] = 0;
  _$jscoverage['util/bignum.js'][28] = 0;
  _$jscoverage['util/bignum.js'][29] = 0;
  _$jscoverage['util/bignum.js'][32] = 0;
  _$jscoverage['util/bignum.js'][33] = 0;
  _$jscoverage['util/bignum.js'][34] = 0;
  _$jscoverage['util/bignum.js'][37] = 0;
  _$jscoverage['util/bignum.js'][38] = 0;
  _$jscoverage['util/bignum.js'][39] = 0;
  _$jscoverage['util/bignum.js'][42] = 0;
  _$jscoverage['util/bignum.js'][43] = 0;
  _$jscoverage['util/bignum.js'][46] = 0;
  _$jscoverage['util/bignum.js'][47] = 0;
  _$jscoverage['util/bignum.js'][48] = 0;
  _$jscoverage['util/bignum.js'][51] = 0;
  _$jscoverage['util/bignum.js'][52] = 0;
  _$jscoverage['util/bignum.js'][53] = 0;
  _$jscoverage['util/bignum.js'][54] = 0;
  _$jscoverage['util/bignum.js'][55] = 0;
  _$jscoverage['util/bignum.js'][58] = 0;
  _$jscoverage['util/bignum.js'][59] = 0;
  _$jscoverage['util/bignum.js'][60] = 0;
  _$jscoverage['util/bignum.js'][61] = 0;
  _$jscoverage['util/bignum.js'][62] = 0;
  _$jscoverage['util/bignum.js'][63] = 0;
  _$jscoverage['util/bignum.js'][66] = 0;
  _$jscoverage['util/bignum.js'][67] = 0;
  _$jscoverage['util/bignum.js'][68] = 0;
  _$jscoverage['util/bignum.js'][69] = 0;
  _$jscoverage['util/bignum.js'][70] = 0;
  _$jscoverage['util/bignum.js'][71] = 0;
  _$jscoverage['util/bignum.js'][72] = 0;
  _$jscoverage['util/bignum.js'][73] = 0;
  _$jscoverage['util/bignum.js'][75] = 0;
  _$jscoverage['util/bignum.js'][77] = 0;
  _$jscoverage['util/bignum.js'][79] = 0;
  _$jscoverage['util/bignum.js'][80] = 0;
  _$jscoverage['util/bignum.js'][81] = 0;
  _$jscoverage['util/bignum.js'][82] = 0;
  _$jscoverage['util/bignum.js'][83] = 0;
  _$jscoverage['util/bignum.js'][85] = 0;
  _$jscoverage['util/bignum.js'][86] = 0;
  _$jscoverage['util/bignum.js'][87] = 0;
  _$jscoverage['util/bignum.js'][88] = 0;
  _$jscoverage['util/bignum.js'][90] = 0;
  _$jscoverage['util/bignum.js'][91] = 0;
  _$jscoverage['util/bignum.js'][92] = 0;
  _$jscoverage['util/bignum.js'][93] = 0;
  _$jscoverage['util/bignum.js'][95] = 0;
  _$jscoverage['util/bignum.js'][97] = 0;
  _$jscoverage['util/bignum.js'][99] = 0;
  _$jscoverage['util/bignum.js'][102] = 0;
  _$jscoverage['util/bignum.js'][103] = 0;
  _$jscoverage['util/bignum.js'][105] = 0;
  _$jscoverage['util/bignum.js'][106] = 0;
  _$jscoverage['util/bignum.js'][109] = 0;
  _$jscoverage['util/bignum.js'][110] = 0;
  _$jscoverage['util/bignum.js'][111] = 0;
  _$jscoverage['util/bignum.js'][113] = 0;
  _$jscoverage['util/bignum.js'][114] = 0;
  _$jscoverage['util/bignum.js'][117] = 0;
  _$jscoverage['util/bignum.js'][118] = 0;
  _$jscoverage['util/bignum.js'][119] = 0;
  _$jscoverage['util/bignum.js'][121] = 0;
  _$jscoverage['util/bignum.js'][122] = 0;
  _$jscoverage['util/bignum.js'][125] = 0;
  _$jscoverage['util/bignum.js'][126] = 0;
  _$jscoverage['util/bignum.js'][127] = 0;
  _$jscoverage['util/bignum.js'][129] = 0;
  _$jscoverage['util/bignum.js'][130] = 0;
  _$jscoverage['util/bignum.js'][131] = 0;
  _$jscoverage['util/bignum.js'][133] = 0;
  _$jscoverage['util/bignum.js'][134] = 0;
  _$jscoverage['util/bignum.js'][135] = 0;
  _$jscoverage['util/bignum.js'][138] = 0;
  _$jscoverage['util/bignum.js'][139] = 0;
  _$jscoverage['util/bignum.js'][141] = 0;
  _$jscoverage['util/bignum.js'][145] = 0;
  _$jscoverage['util/bignum.js'][146] = 0;
  _$jscoverage['util/bignum.js'][147] = 0;
  _$jscoverage['util/bignum.js'][149] = 0;
  _$jscoverage['util/bignum.js'][150] = 0;
  _$jscoverage['util/bignum.js'][152] = 0;
  _$jscoverage['util/bignum.js'][153] = 0;
  _$jscoverage['util/bignum.js'][155] = 0;
  _$jscoverage['util/bignum.js'][158] = 0;
  _$jscoverage['util/bignum.js'][159] = 0;
  _$jscoverage['util/bignum.js'][160] = 0;
  _$jscoverage['util/bignum.js'][162] = 0;
  _$jscoverage['util/bignum.js'][163] = 0;
  _$jscoverage['util/bignum.js'][165] = 0;
  _$jscoverage['util/bignum.js'][168] = 0;
  _$jscoverage['util/bignum.js'][170] = 0;
  _$jscoverage['util/bignum.js'][171] = 0;
  _$jscoverage['util/bignum.js'][172] = 0;
  _$jscoverage['util/bignum.js'][173] = 0;
  _$jscoverage['util/bignum.js'][175] = 0;
  _$jscoverage['util/bignum.js'][176] = 0;
  _$jscoverage['util/bignum.js'][177] = 0;
  _$jscoverage['util/bignum.js'][179] = 0;
  _$jscoverage['util/bignum.js'][182] = 0;
  _$jscoverage['util/bignum.js'][184] = 0;
  _$jscoverage['util/bignum.js'][186] = 0;
  _$jscoverage['util/bignum.js'][188] = 0;
  _$jscoverage['util/bignum.js'][189] = 0;
  _$jscoverage['util/bignum.js'][192] = 0;
  _$jscoverage['util/bignum.js'][193] = 0;
  _$jscoverage['util/bignum.js'][194] = 0;
  _$jscoverage['util/bignum.js'][195] = 0;
  _$jscoverage['util/bignum.js'][196] = 0;
  _$jscoverage['util/bignum.js'][198] = 0;
  _$jscoverage['util/bignum.js'][199] = 0;
  _$jscoverage['util/bignum.js'][200] = 0;
  _$jscoverage['util/bignum.js'][201] = 0;
  _$jscoverage['util/bignum.js'][202] = 0;
  _$jscoverage['util/bignum.js'][204] = 0;
  _$jscoverage['util/bignum.js'][205] = 0;
  _$jscoverage['util/bignum.js'][206] = 0;
  _$jscoverage['util/bignum.js'][207] = 0;
  _$jscoverage['util/bignum.js'][208] = 0;
  _$jscoverage['util/bignum.js'][210] = 0;
  _$jscoverage['util/bignum.js'][211] = 0;
  _$jscoverage['util/bignum.js'][212] = 0;
  _$jscoverage['util/bignum.js'][213] = 0;
  _$jscoverage['util/bignum.js'][215] = 0;
  _$jscoverage['util/bignum.js'][222] = 0;
  _$jscoverage['util/bignum.js'][223] = 0;
  _$jscoverage['util/bignum.js'][224] = 0;
  _$jscoverage['util/bignum.js'][225] = 0;
  _$jscoverage['util/bignum.js'][227] = 0;
  _$jscoverage['util/bignum.js'][228] = 0;
  _$jscoverage['util/bignum.js'][229] = 0;
  _$jscoverage['util/bignum.js'][233] = 0;
  _$jscoverage['util/bignum.js'][234] = 0;
  _$jscoverage['util/bignum.js'][235] = 0;
  _$jscoverage['util/bignum.js'][237] = 0;
  _$jscoverage['util/bignum.js'][238] = 0;
  _$jscoverage['util/bignum.js'][241] = 0;
  _$jscoverage['util/bignum.js'][242] = 0;
  _$jscoverage['util/bignum.js'][243] = 0;
  _$jscoverage['util/bignum.js'][244] = 0;
  _$jscoverage['util/bignum.js'][246] = 0;
  _$jscoverage['util/bignum.js'][247] = 0;
  _$jscoverage['util/bignum.js'][248] = 0;
  _$jscoverage['util/bignum.js'][249] = 0;
  _$jscoverage['util/bignum.js'][250] = 0;
  _$jscoverage['util/bignum.js'][253] = 0;
  _$jscoverage['util/bignum.js'][254] = 0;
  _$jscoverage['util/bignum.js'][255] = 0;
  _$jscoverage['util/bignum.js'][257] = 0;
  _$jscoverage['util/bignum.js'][258] = 0;
  _$jscoverage['util/bignum.js'][261] = 0;
  _$jscoverage['util/bignum.js'][262] = 0;
  _$jscoverage['util/bignum.js'][263] = 0;
  _$jscoverage['util/bignum.js'][264] = 0;
  _$jscoverage['util/bignum.js'][266] = 0;
  _$jscoverage['util/bignum.js'][267] = 0;
  _$jscoverage['util/bignum.js'][268] = 0;
  _$jscoverage['util/bignum.js'][269] = 0;
  _$jscoverage['util/bignum.js'][270] = 0;
  _$jscoverage['util/bignum.js'][273] = 0;
  _$jscoverage['util/bignum.js'][274] = 0;
  _$jscoverage['util/bignum.js'][275] = 0;
  _$jscoverage['util/bignum.js'][277] = 0;
  _$jscoverage['util/bignum.js'][278] = 0;
  _$jscoverage['util/bignum.js'][279] = 0;
  _$jscoverage['util/bignum.js'][282] = 0;
  _$jscoverage['util/bignum.js'][283] = 0;
  _$jscoverage['util/bignum.js'][284] = 0;
  _$jscoverage['util/bignum.js'][285] = 0;
  _$jscoverage['util/bignum.js'][286] = 0;
  _$jscoverage['util/bignum.js'][288] = 0;
  _$jscoverage['util/bignum.js'][289] = 0;
  _$jscoverage['util/bignum.js'][290] = 0;
  _$jscoverage['util/bignum.js'][292] = 0;
  _$jscoverage['util/bignum.js'][293] = 0;
  _$jscoverage['util/bignum.js'][294] = 0;
  _$jscoverage['util/bignum.js'][296] = 0;
  _$jscoverage['util/bignum.js'][297] = 0;
  _$jscoverage['util/bignum.js'][298] = 0;
  _$jscoverage['util/bignum.js'][300] = 0;
  _$jscoverage['util/bignum.js'][301] = 0;
  _$jscoverage['util/bignum.js'][302] = 0;
  _$jscoverage['util/bignum.js'][305] = 0;
  _$jscoverage['util/bignum.js'][306] = 0;
  _$jscoverage['util/bignum.js'][308] = 0;
  _$jscoverage['util/bignum.js'][310] = 0;
  _$jscoverage['util/bignum.js'][311] = 0;
  _$jscoverage['util/bignum.js'][312] = 0;
  _$jscoverage['util/bignum.js'][314] = 0;
  _$jscoverage['util/bignum.js'][315] = 0;
  _$jscoverage['util/bignum.js'][317] = 0;
  _$jscoverage['util/bignum.js'][318] = 0;
  _$jscoverage['util/bignum.js'][320] = 0;
  _$jscoverage['util/bignum.js'][321] = 0;
  _$jscoverage['util/bignum.js'][323] = 0;
  _$jscoverage['util/bignum.js'][324] = 0;
  _$jscoverage['util/bignum.js'][328] = 0;
  _$jscoverage['util/bignum.js'][329] = 0;
  _$jscoverage['util/bignum.js'][330] = 0;
  _$jscoverage['util/bignum.js'][331] = 0;
  _$jscoverage['util/bignum.js'][333] = 0;
  _$jscoverage['util/bignum.js'][334] = 0;
  _$jscoverage['util/bignum.js'][335] = 0;
  _$jscoverage['util/bignum.js'][336] = 0;
  _$jscoverage['util/bignum.js'][337] = 0;
  _$jscoverage['util/bignum.js'][338] = 0;
  _$jscoverage['util/bignum.js'][340] = 0;
  _$jscoverage['util/bignum.js'][342] = 0;
  _$jscoverage['util/bignum.js'][343] = 0;
  _$jscoverage['util/bignum.js'][345] = 0;
  _$jscoverage['util/bignum.js'][346] = 0;
  _$jscoverage['util/bignum.js'][348] = 0;
  _$jscoverage['util/bignum.js'][349] = 0;
  _$jscoverage['util/bignum.js'][351] = 0;
  _$jscoverage['util/bignum.js'][354] = 0;
  _$jscoverage['util/bignum.js'][355] = 0;
  _$jscoverage['util/bignum.js'][356] = 0;
  _$jscoverage['util/bignum.js'][357] = 0;
  _$jscoverage['util/bignum.js'][359] = 0;
  _$jscoverage['util/bignum.js'][360] = 0;
  _$jscoverage['util/bignum.js'][361] = 0;
  _$jscoverage['util/bignum.js'][362] = 0;
  _$jscoverage['util/bignum.js'][363] = 0;
  _$jscoverage['util/bignum.js'][364] = 0;
  _$jscoverage['util/bignum.js'][365] = 0;
  _$jscoverage['util/bignum.js'][366] = 0;
  _$jscoverage['util/bignum.js'][367] = 0;
  _$jscoverage['util/bignum.js'][368] = 0;
  _$jscoverage['util/bignum.js'][369] = 0;
  _$jscoverage['util/bignum.js'][370] = 0;
  _$jscoverage['util/bignum.js'][371] = 0;
  _$jscoverage['util/bignum.js'][373] = 0;
  _$jscoverage['util/bignum.js'][374] = 0;
  _$jscoverage['util/bignum.js'][376] = 0;
  _$jscoverage['util/bignum.js'][377] = 0;
  _$jscoverage['util/bignum.js'][378] = 0;
  _$jscoverage['util/bignum.js'][380] = 0;
  _$jscoverage['util/bignum.js'][381] = 0;
  _$jscoverage['util/bignum.js'][383] = 0;
  _$jscoverage['util/bignum.js'][384] = 0;
  _$jscoverage['util/bignum.js'][386] = 0;
  _$jscoverage['util/bignum.js'][387] = 0;
  _$jscoverage['util/bignum.js'][389] = 0;
  _$jscoverage['util/bignum.js'][390] = 0;
  _$jscoverage['util/bignum.js'][391] = 0;
  _$jscoverage['util/bignum.js'][392] = 0;
  _$jscoverage['util/bignum.js'][393] = 0;
  _$jscoverage['util/bignum.js'][395] = 0;
  _$jscoverage['util/bignum.js'][399] = 0;
  _$jscoverage['util/bignum.js'][401] = 0;
  _$jscoverage['util/bignum.js'][403] = 0;
  _$jscoverage['util/bignum.js'][405] = 0;
  _$jscoverage['util/bignum.js'][407] = 0;
  _$jscoverage['util/bignum.js'][409] = 0;
  _$jscoverage['util/bignum.js'][411] = 0;
  _$jscoverage['util/bignum.js'][412] = 0;
  _$jscoverage['util/bignum.js'][413] = 0;
  _$jscoverage['util/bignum.js'][415] = 0;
  _$jscoverage['util/bignum.js'][416] = 0;
  _$jscoverage['util/bignum.js'][417] = 0;
  _$jscoverage['util/bignum.js'][420] = 0;
  _$jscoverage['util/bignum.js'][421] = 0;
  _$jscoverage['util/bignum.js'][422] = 0;
  _$jscoverage['util/bignum.js'][424] = 0;
  _$jscoverage['util/bignum.js'][425] = 0;
  _$jscoverage['util/bignum.js'][427] = 0;
  _$jscoverage['util/bignum.js'][428] = 0;
  _$jscoverage['util/bignum.js'][429] = 0;
  _$jscoverage['util/bignum.js'][431] = 0;
  _$jscoverage['util/bignum.js'][432] = 0;
  _$jscoverage['util/bignum.js'][434] = 0;
  _$jscoverage['util/bignum.js'][435] = 0;
  _$jscoverage['util/bignum.js'][436] = 0;
  _$jscoverage['util/bignum.js'][437] = 0;
  _$jscoverage['util/bignum.js'][439] = 0;
  _$jscoverage['util/bignum.js'][440] = 0;
  _$jscoverage['util/bignum.js'][441] = 0;
  _$jscoverage['util/bignum.js'][443] = 0;
  _$jscoverage['util/bignum.js'][444] = 0;
  _$jscoverage['util/bignum.js'][445] = 0;
  _$jscoverage['util/bignum.js'][447] = 0;
  _$jscoverage['util/bignum.js'][448] = 0;
  _$jscoverage['util/bignum.js'][449] = 0;
  _$jscoverage['util/bignum.js'][452] = 0;
  _$jscoverage['util/bignum.js'][453] = 0;
  _$jscoverage['util/bignum.js'][456] = 0;
  _$jscoverage['util/bignum.js'][458] = 0;
  _$jscoverage['util/bignum.js'][459] = 0;
  _$jscoverage['util/bignum.js'][460] = 0;
  _$jscoverage['util/bignum.js'][463] = 0;
  _$jscoverage['util/bignum.js'][464] = 0;
  _$jscoverage['util/bignum.js'][467] = 0;
  _$jscoverage['util/bignum.js'][469] = 0;
  _$jscoverage['util/bignum.js'][470] = 0;
  _$jscoverage['util/bignum.js'][471] = 0;
  _$jscoverage['util/bignum.js'][474] = 0;
  _$jscoverage['util/bignum.js'][475] = 0;
  _$jscoverage['util/bignum.js'][478] = 0;
  _$jscoverage['util/bignum.js'][479] = 0;
  _$jscoverage['util/bignum.js'][480] = 0;
  _$jscoverage['util/bignum.js'][481] = 0;
  _$jscoverage['util/bignum.js'][483] = 0;
  _$jscoverage['util/bignum.js'][487] = 0;
  _$jscoverage['util/bignum.js'][488] = 0;
  _$jscoverage['util/bignum.js'][489] = 0;
  _$jscoverage['util/bignum.js'][490] = 0;
  _$jscoverage['util/bignum.js'][491] = 0;
  _$jscoverage['util/bignum.js'][493] = 0;
  _$jscoverage['util/bignum.js'][494] = 0;
  _$jscoverage['util/bignum.js'][495] = 0;
  _$jscoverage['util/bignum.js'][496] = 0;
  _$jscoverage['util/bignum.js'][499] = 0;
  _$jscoverage['util/bignum.js'][501] = 0;
  _$jscoverage['util/bignum.js'][503] = 0;
  _$jscoverage['util/bignum.js'][505] = 0;
  _$jscoverage['util/bignum.js'][506] = 0;
  _$jscoverage['util/bignum.js'][507] = 0;
  _$jscoverage['util/bignum.js'][508] = 0;
  _$jscoverage['util/bignum.js'][509] = 0;
  _$jscoverage['util/bignum.js'][511] = 0;
  _$jscoverage['util/bignum.js'][512] = 0;
  _$jscoverage['util/bignum.js'][513] = 0;
  _$jscoverage['util/bignum.js'][514] = 0;
  _$jscoverage['util/bignum.js'][515] = 0;
  _$jscoverage['util/bignum.js'][517] = 0;
  _$jscoverage['util/bignum.js'][518] = 0;
  _$jscoverage['util/bignum.js'][519] = 0;
  _$jscoverage['util/bignum.js'][520] = 0;
  _$jscoverage['util/bignum.js'][521] = 0;
  _$jscoverage['util/bignum.js'][523] = 0;
  _$jscoverage['util/bignum.js'][524] = 0;
  _$jscoverage['util/bignum.js'][525] = 0;
  _$jscoverage['util/bignum.js'][526] = 0;
  _$jscoverage['util/bignum.js'][528] = 0;
  _$jscoverage['util/bignum.js'][529] = 0;
  _$jscoverage['util/bignum.js'][530] = 0;
  _$jscoverage['util/bignum.js'][532] = 0;
  _$jscoverage['util/bignum.js'][533] = 0;
  _$jscoverage['util/bignum.js'][537] = 0;
  _$jscoverage['util/bignum.js'][538] = 0;
  _$jscoverage['util/bignum.js'][539] = 0;
  _$jscoverage['util/bignum.js'][541] = 0;
  _$jscoverage['util/bignum.js'][542] = 0;
  _$jscoverage['util/bignum.js'][545] = 0;
  _$jscoverage['util/bignum.js'][546] = 0;
  _$jscoverage['util/bignum.js'][547] = 0;
  _$jscoverage['util/bignum.js'][548] = 0;
  _$jscoverage['util/bignum.js'][550] = 0;
  _$jscoverage['util/bignum.js'][551] = 0;
  _$jscoverage['util/bignum.js'][552] = 0;
  _$jscoverage['util/bignum.js'][553] = 0;
  _$jscoverage['util/bignum.js'][554] = 0;
  _$jscoverage['util/bignum.js'][557] = 0;
  _$jscoverage['util/bignum.js'][558] = 0;
  _$jscoverage['util/bignum.js'][559] = 0;
  _$jscoverage['util/bignum.js'][561] = 0;
  _$jscoverage['util/bignum.js'][562] = 0;
  _$jscoverage['util/bignum.js'][565] = 0;
  _$jscoverage['util/bignum.js'][566] = 0;
  _$jscoverage['util/bignum.js'][567] = 0;
  _$jscoverage['util/bignum.js'][568] = 0;
  _$jscoverage['util/bignum.js'][570] = 0;
  _$jscoverage['util/bignum.js'][571] = 0;
  _$jscoverage['util/bignum.js'][572] = 0;
  _$jscoverage['util/bignum.js'][573] = 0;
  _$jscoverage['util/bignum.js'][574] = 0;
  _$jscoverage['util/bignum.js'][577] = 0;
  _$jscoverage['util/bignum.js'][578] = 0;
  _$jscoverage['util/bignum.js'][579] = 0;
  _$jscoverage['util/bignum.js'][581] = 0;
  _$jscoverage['util/bignum.js'][582] = 0;
  _$jscoverage['util/bignum.js'][583] = 0;
  _$jscoverage['util/bignum.js'][586] = 0;
  _$jscoverage['util/bignum.js'][587] = 0;
  _$jscoverage['util/bignum.js'][588] = 0;
  _$jscoverage['util/bignum.js'][589] = 0;
  _$jscoverage['util/bignum.js'][590] = 0;
  _$jscoverage['util/bignum.js'][592] = 0;
  _$jscoverage['util/bignum.js'][593] = 0;
  _$jscoverage['util/bignum.js'][594] = 0;
  _$jscoverage['util/bignum.js'][596] = 0;
  _$jscoverage['util/bignum.js'][597] = 0;
  _$jscoverage['util/bignum.js'][598] = 0;
  _$jscoverage['util/bignum.js'][600] = 0;
  _$jscoverage['util/bignum.js'][601] = 0;
  _$jscoverage['util/bignum.js'][602] = 0;
  _$jscoverage['util/bignum.js'][604] = 0;
  _$jscoverage['util/bignum.js'][605] = 0;
  _$jscoverage['util/bignum.js'][606] = 0;
  _$jscoverage['util/bignum.js'][609] = 0;
  _$jscoverage['util/bignum.js'][610] = 0;
  _$jscoverage['util/bignum.js'][613] = 0;
  _$jscoverage['util/bignum.js'][614] = 0;
  _$jscoverage['util/bignum.js'][616] = 0;
  _$jscoverage['util/bignum.js'][617] = 0;
  _$jscoverage['util/bignum.js'][619] = 0;
  _$jscoverage['util/bignum.js'][621] = 0;
  _$jscoverage['util/bignum.js'][624] = 0;
  _$jscoverage['util/bignum.js'][628] = 0;
  _$jscoverage['util/bignum.js'][630] = 0;
  _$jscoverage['util/bignum.js'][633] = 0;
  _$jscoverage['util/bignum.js'][635] = 0;
  _$jscoverage['util/bignum.js'][636] = 0;
  _$jscoverage['util/bignum.js'][637] = 0;
  _$jscoverage['util/bignum.js'][639] = 0;
  _$jscoverage['util/bignum.js'][641] = 0;
  _$jscoverage['util/bignum.js'][642] = 0;
  _$jscoverage['util/bignum.js'][644] = 0;
  _$jscoverage['util/bignum.js'][646] = 0;
  _$jscoverage['util/bignum.js'][647] = 0;
  _$jscoverage['util/bignum.js'][649] = 0;
  _$jscoverage['util/bignum.js'][651] = 0;
  _$jscoverage['util/bignum.js'][652] = 0;
  _$jscoverage['util/bignum.js'][654] = 0;
  _$jscoverage['util/bignum.js'][656] = 0;
  _$jscoverage['util/bignum.js'][657] = 0;
  _$jscoverage['util/bignum.js'][659] = 0;
  _$jscoverage['util/bignum.js'][662] = 0;
  _$jscoverage['util/bignum.js'][665] = 0;
  _$jscoverage['util/bignum.js'][666] = 0;
  _$jscoverage['util/bignum.js'][667] = 0;
  _$jscoverage['util/bignum.js'][671] = 0;
  _$jscoverage['util/bignum.js'][672] = 0;
  _$jscoverage['util/bignum.js'][673] = 0;
  _$jscoverage['util/bignum.js'][674] = 0;
  _$jscoverage['util/bignum.js'][675] = 0;
  _$jscoverage['util/bignum.js'][676] = 0;
  _$jscoverage['util/bignum.js'][680] = 0;
  _$jscoverage['util/bignum.js'][681] = 0;
  _$jscoverage['util/bignum.js'][682] = 0;
  _$jscoverage['util/bignum.js'][683] = 0;
  _$jscoverage['util/bignum.js'][684] = 0;
  _$jscoverage['util/bignum.js'][685] = 0;
  _$jscoverage['util/bignum.js'][686] = 0;
  _$jscoverage['util/bignum.js'][690] = 0;
  _$jscoverage['util/bignum.js'][691] = 0;
  _$jscoverage['util/bignum.js'][692] = 0;
  _$jscoverage['util/bignum.js'][693] = 0;
  _$jscoverage['util/bignum.js'][694] = 0;
  _$jscoverage['util/bignum.js'][695] = 0;
  _$jscoverage['util/bignum.js'][696] = 0;
  _$jscoverage['util/bignum.js'][697] = 0;
  _$jscoverage['util/bignum.js'][698] = 0;
  _$jscoverage['util/bignum.js'][702] = 0;
  _$jscoverage['util/bignum.js'][703] = 0;
  _$jscoverage['util/bignum.js'][704] = 0;
  _$jscoverage['util/bignum.js'][705] = 0;
  _$jscoverage['util/bignum.js'][706] = 0;
  _$jscoverage['util/bignum.js'][707] = 0;
  _$jscoverage['util/bignum.js'][708] = 0;
  _$jscoverage['util/bignum.js'][709] = 0;
  _$jscoverage['util/bignum.js'][710] = 0;
  _$jscoverage['util/bignum.js'][711] = 0;
  _$jscoverage['util/bignum.js'][715] = 0;
  _$jscoverage['util/bignum.js'][716] = 0;
  _$jscoverage['util/bignum.js'][717] = 0;
  _$jscoverage['util/bignum.js'][718] = 0;
  _$jscoverage['util/bignum.js'][719] = 0;
  _$jscoverage['util/bignum.js'][720] = 0;
  _$jscoverage['util/bignum.js'][721] = 0;
  _$jscoverage['util/bignum.js'][722] = 0;
  _$jscoverage['util/bignum.js'][723] = 0;
  _$jscoverage['util/bignum.js'][724] = 0;
  _$jscoverage['util/bignum.js'][727] = 0;
  _$jscoverage['util/bignum.js'][728] = 0;
  _$jscoverage['util/bignum.js'][729] = 0;
  _$jscoverage['util/bignum.js'][731] = 0;
  _$jscoverage['util/bignum.js'][732] = 0;
  _$jscoverage['util/bignum.js'][733] = 0;
  _$jscoverage['util/bignum.js'][735] = 0;
  _$jscoverage['util/bignum.js'][736] = 0;
  _$jscoverage['util/bignum.js'][737] = 0;
  _$jscoverage['util/bignum.js'][739] = 0;
  _$jscoverage['util/bignum.js'][740] = 0;
  _$jscoverage['util/bignum.js'][741] = 0;
  _$jscoverage['util/bignum.js'][743] = 0;
  _$jscoverage['util/bignum.js'][744] = 0;
  _$jscoverage['util/bignum.js'][745] = 0;
  _$jscoverage['util/bignum.js'][747] = 0;
  _$jscoverage['util/bignum.js'][748] = 0;
  _$jscoverage['util/bignum.js'][749] = 0;
  _$jscoverage['util/bignum.js'][751] = 0;
  _$jscoverage['util/bignum.js'][752] = 0;
  _$jscoverage['util/bignum.js'][753] = 0;
  _$jscoverage['util/bignum.js'][756] = 0;
  _$jscoverage['util/bignum.js'][757] = 0;
  _$jscoverage['util/bignum.js'][758] = 0;
  _$jscoverage['util/bignum.js'][759] = 0;
  _$jscoverage['util/bignum.js'][761] = 0;
  _$jscoverage['util/bignum.js'][762] = 0;
  _$jscoverage['util/bignum.js'][763] = 0;
  _$jscoverage['util/bignum.js'][764] = 0;
  _$jscoverage['util/bignum.js'][765] = 0;
  _$jscoverage['util/bignum.js'][767] = 0;
  _$jscoverage['util/bignum.js'][768] = 0;
  _$jscoverage['util/bignum.js'][769] = 0;
  _$jscoverage['util/bignum.js'][770] = 0;
  _$jscoverage['util/bignum.js'][771] = 0;
  _$jscoverage['util/bignum.js'][773] = 0;
  _$jscoverage['util/bignum.js'][774] = 0;
  _$jscoverage['util/bignum.js'][775] = 0;
  _$jscoverage['util/bignum.js'][776] = 0;
  _$jscoverage['util/bignum.js'][777] = 0;
  _$jscoverage['util/bignum.js'][779] = 0;
  _$jscoverage['util/bignum.js'][780] = 0;
  _$jscoverage['util/bignum.js'][781] = 0;
  _$jscoverage['util/bignum.js'][782] = 0;
  _$jscoverage['util/bignum.js'][785] = 0;
  _$jscoverage['util/bignum.js'][786] = 0;
  _$jscoverage['util/bignum.js'][789] = 0;
  _$jscoverage['util/bignum.js'][790] = 0;
  _$jscoverage['util/bignum.js'][793] = 0;
  _$jscoverage['util/bignum.js'][794] = 0;
  _$jscoverage['util/bignum.js'][797] = 0;
  _$jscoverage['util/bignum.js'][798] = 0;
  _$jscoverage['util/bignum.js'][801] = 0;
  _$jscoverage['util/bignum.js'][802] = 0;
  _$jscoverage['util/bignum.js'][803] = 0;
  _$jscoverage['util/bignum.js'][804] = 0;
  _$jscoverage['util/bignum.js'][805] = 0;
  _$jscoverage['util/bignum.js'][806] = 0;
  _$jscoverage['util/bignum.js'][807] = 0;
  _$jscoverage['util/bignum.js'][808] = 0;
  _$jscoverage['util/bignum.js'][809] = 0;
  _$jscoverage['util/bignum.js'][810] = 0;
}
_$jscoverage['util/bignum.js'][14]++;
"use strict";
_$jscoverage['util/bignum.js'][17]++;
var zeropad = require("./zeropad");
_$jscoverage['util/bignum.js'][19]++;
var BASE = Math.pow(10, 7);
_$jscoverage['util/bignum.js'][21]++;
var EXP_BIAS = 6176;
_$jscoverage['util/bignum.js'][23]++;
var INT_2_16 = Math.pow(2, 16);
_$jscoverage['util/bignum.js'][24]++;
var INT_2_21 = Math.pow(2, 21);
_$jscoverage['util/bignum.js'][25]++;
var INT_2_32 = Math.pow(2, 32);
_$jscoverage['util/bignum.js'][28]++;
var INT_2_32_0 = 4967296;
_$jscoverage['util/bignum.js'][29]++;
var INT_2_32_1 = 429;
_$jscoverage['util/bignum.js'][32]++;
var INT_2_53_0 = 4740992;
_$jscoverage['util/bignum.js'][33]++;
var INT_2_53_1 = 719925;
_$jscoverage['util/bignum.js'][34]++;
var INT_2_53_2 = 90;
_$jscoverage['util/bignum.js'][37]++;
var INT_2_64_0 = 9551616;
_$jscoverage['util/bignum.js'][38]++;
var INT_2_64_1 = 4407370;
_$jscoverage['util/bignum.js'][39]++;
var INT_2_64_2 = 184467;
_$jscoverage['util/bignum.js'][42]++;
var BIN_10_7_0 = 38528;
_$jscoverage['util/bignum.js'][43]++;
var BIN_10_7_1 = 152;
_$jscoverage['util/bignum.js'][46]++;
var BIN_10_14_0 = 16384;
_$jscoverage['util/bignum.js'][47]++;
var BIN_10_14_1 = 4218;
_$jscoverage['util/bignum.js'][48]++;
var BIN_10_14_2 = 23283;
_$jscoverage['util/bignum.js'][51]++;
var BIN_10_21_0 = 0;
_$jscoverage['util/bignum.js'][52]++;
var BIN_10_21_1 = 56992;
_$jscoverage['util/bignum.js'][53]++;
var BIN_10_21_2 = 44485;
_$jscoverage['util/bignum.js'][54]++;
var BIN_10_21_3 = 13769;
_$jscoverage['util/bignum.js'][55]++;
var BIN_10_21_4 = 54;
_$jscoverage['util/bignum.js'][58]++;
var BIN_10_28_0 = 0;
_$jscoverage['util/bignum.js'][59]++;
var BIN_10_28_1 = 4096;
_$jscoverage['util/bignum.js'][60]++;
var BIN_10_28_2 = 609;
_$jscoverage['util/bignum.js'][61]++;
var BIN_10_28_3 = 15909;
_$jscoverage['util/bignum.js'][62]++;
var BIN_10_28_4 = 52830;
_$jscoverage['util/bignum.js'][63]++;
var BIN_10_28_5 = 8271;
_$jscoverage['util/bignum.js'][66]++;
var BIN_10_35_0 = 0;
_$jscoverage['util/bignum.js'][67]++;
var BIN_10_35_1 = 0;
_$jscoverage['util/bignum.js'][68]++;
var BIN_10_35_2 = 36840;
_$jscoverage['util/bignum.js'][69]++;
var BIN_10_35_3 = 11143;
_$jscoverage['util/bignum.js'][70]++;
var BIN_10_35_4 = 19842;
_$jscoverage['util/bignum.js'][71]++;
var BIN_10_35_5 = 29383;
_$jscoverage['util/bignum.js'][72]++;
var BIN_10_35_6 = 16993;
_$jscoverage['util/bignum.js'][73]++;
var BIN_10_35_7 = 19;
_$jscoverage['util/bignum.js'][75]++;
function _readInt64(buffer, offset, unsigned) {
  _$jscoverage['util/bignum.js'][77]++;
  var x, y, s, y0, y1, x0, x1, x2;
  _$jscoverage['util/bignum.js'][79]++;
  x = buffer[offset + 2] << 16;
  _$jscoverage['util/bignum.js'][80]++;
  x |= buffer[offset + 1] << 8;
  _$jscoverage['util/bignum.js'][81]++;
  x |= buffer[offset];
  _$jscoverage['util/bignum.js'][82]++;
  x += buffer[offset + 3] << 24 >>> 0;
  _$jscoverage['util/bignum.js'][83]++;
  offset += 4;
  _$jscoverage['util/bignum.js'][85]++;
  y = buffer[offset + 2] << 16;
  _$jscoverage['util/bignum.js'][86]++;
  y |= buffer[offset + 1] << 8;
  _$jscoverage['util/bignum.js'][87]++;
  y |= buffer[offset];
  _$jscoverage['util/bignum.js'][88]++;
  y += buffer[offset + 3] << 24 >>> 0;
  _$jscoverage['util/bignum.js'][90]++;
  if (! unsigned && (y & 2147483648)) {
    _$jscoverage['util/bignum.js'][91]++;
    y = 4294967295 - y;
    _$jscoverage['util/bignum.js'][92]++;
    if (x === 0) {
      _$jscoverage['util/bignum.js'][93]++;
      y += 1;
    }
    else {
      _$jscoverage['util/bignum.js'][95]++;
      x = 4294967295 - x + 1;
    }
    _$jscoverage['util/bignum.js'][97]++;
    s = -1;
  }
  else {
    _$jscoverage['util/bignum.js'][99]++;
    s = 1;
  }
  _$jscoverage['util/bignum.js'][102]++;
  if (y === 0) {
    _$jscoverage['util/bignum.js'][103]++;
    return s * x;
  }
  _$jscoverage['util/bignum.js'][105]++;
  if (y < INT_2_21 || (y === INT_2_21 && x === 0)) {
    _$jscoverage['util/bignum.js'][106]++;
    return s * (y * INT_2_32 + x);
  }
  _$jscoverage['util/bignum.js'][109]++;
  if (x < BASE) {
    _$jscoverage['util/bignum.js'][110]++;
    x0 = x % BASE;
    _$jscoverage['util/bignum.js'][111]++;
    x1 = 0;
  }
  else {
    _$jscoverage['util/bignum.js'][113]++;
    x0 = x % BASE;
    _$jscoverage['util/bignum.js'][114]++;
    x1 = Math.floor(x / BASE);
  }
  _$jscoverage['util/bignum.js'][117]++;
  if (y < BASE) {
    _$jscoverage['util/bignum.js'][118]++;
    y0 = y % BASE;
    _$jscoverage['util/bignum.js'][119]++;
    y1 = 0;
  }
  else {
    _$jscoverage['util/bignum.js'][121]++;
    y0 = y % BASE;
    _$jscoverage['util/bignum.js'][122]++;
    y1 = Math.floor(y / BASE);
  }
  _$jscoverage['util/bignum.js'][125]++;
  x0 += y0 * INT_2_32_0;
  _$jscoverage['util/bignum.js'][126]++;
  x1 += y0 * INT_2_32_1 + y1 * INT_2_32_0;
  _$jscoverage['util/bignum.js'][127]++;
  x2 = y1 * INT_2_32_1;
  _$jscoverage['util/bignum.js'][129]++;
  if (x0 >= BASE) {
    _$jscoverage['util/bignum.js'][130]++;
    x1 += Math.floor(x0 / BASE);
    _$jscoverage['util/bignum.js'][131]++;
    x0 %= BASE;
  }
  _$jscoverage['util/bignum.js'][133]++;
  if (x1 >= BASE) {
    _$jscoverage['util/bignum.js'][134]++;
    x2 += Math.floor(x1 / BASE);
    _$jscoverage['util/bignum.js'][135]++;
    x1 %= BASE;
  }
  _$jscoverage['util/bignum.js'][138]++;
  if (s === 1) {
    _$jscoverage['util/bignum.js'][139]++;
    return "" + x2 + zeropad.lpad14(x1 * BASE + x0);
  }
  _$jscoverage['util/bignum.js'][141]++;
  return "-" + x2 + zeropad.lpad14(x1 * BASE + x0);
}
_$jscoverage['util/bignum.js'][145]++;
function isUInt64(value) {
  _$jscoverage['util/bignum.js'][146]++;
  if (typeof value === "number") {
    _$jscoverage['util/bignum.js'][147]++;
    return true;
  }
  _$jscoverage['util/bignum.js'][149]++;
  if (value.length < 20) {
    _$jscoverage['util/bignum.js'][150]++;
    return true;
  }
  _$jscoverage['util/bignum.js'][152]++;
  if (value.length === 20 && value <= "18446744073709551616") {
    _$jscoverage['util/bignum.js'][153]++;
    return true;
  }
  _$jscoverage['util/bignum.js'][155]++;
  return false;
}
_$jscoverage['util/bignum.js'][158]++;
function writeDec128(buffer, value, offset) {
  _$jscoverage['util/bignum.js'][159]++;
  offset = offset || 0;
  _$jscoverage['util/bignum.js'][160]++;
  buffer.fill(0, offset, offset + 16);
  _$jscoverage['util/bignum.js'][162]++;
  if (isUInt64(value.m)) {
    _$jscoverage['util/bignum.js'][163]++;
    writeUInt64(buffer, value.m, offset);
  }
  else {
    _$jscoverage['util/bignum.js'][165]++;
    writeUInt128(buffer, value.m, offset);
  }
  _$jscoverage['util/bignum.js'][168]++;
  var e = EXP_BIAS + value.e;
  _$jscoverage['util/bignum.js'][170]++;
  var e0 = e << 1;
  _$jscoverage['util/bignum.js'][171]++;
  e0 &= 254;
  _$jscoverage['util/bignum.js'][172]++;
  e0 |= buffer[offset + 14] & 1;
  _$jscoverage['util/bignum.js'][173]++;
  buffer[offset + 14] = e0;
  _$jscoverage['util/bignum.js'][175]++;
  var e1 = e >> 7;
  _$jscoverage['util/bignum.js'][176]++;
  if (value.s < 0) {
    _$jscoverage['util/bignum.js'][177]++;
    e1 |= 128;
  }
  _$jscoverage['util/bignum.js'][179]++;
  buffer[offset + 15] = e1;
}
_$jscoverage['util/bignum.js'][182]++;
function readDec128(buffer, offset) {
  _$jscoverage['util/bignum.js'][184]++;
  var i, j, k, l, z0, z1, y0, y1, y2, x0, x1, x2, x3, x4;
  _$jscoverage['util/bignum.js'][186]++;
  offset = offset || 0;
  _$jscoverage['util/bignum.js'][188]++;
  if ((buffer[offset + 15] & 112) === 112) {
    _$jscoverage['util/bignum.js'][189]++;
    return null;
  }
  _$jscoverage['util/bignum.js'][192]++;
  i = buffer[offset + 2] << 16;
  _$jscoverage['util/bignum.js'][193]++;
  i |= buffer[offset + 1] << 8;
  _$jscoverage['util/bignum.js'][194]++;
  i |= buffer[offset];
  _$jscoverage['util/bignum.js'][195]++;
  i += buffer[offset + 3] << 24 >>> 0;
  _$jscoverage['util/bignum.js'][196]++;
  offset += 4;
  _$jscoverage['util/bignum.js'][198]++;
  j = buffer[offset + 2] << 16;
  _$jscoverage['util/bignum.js'][199]++;
  j |= buffer[offset + 1] << 8;
  _$jscoverage['util/bignum.js'][200]++;
  j |= buffer[offset];
  _$jscoverage['util/bignum.js'][201]++;
  j += buffer[offset + 3] << 24 >>> 0;
  _$jscoverage['util/bignum.js'][202]++;
  offset += 4;
  _$jscoverage['util/bignum.js'][204]++;
  k = buffer[offset + 2] << 16;
  _$jscoverage['util/bignum.js'][205]++;
  k |= buffer[offset + 1] << 8;
  _$jscoverage['util/bignum.js'][206]++;
  k |= buffer[offset];
  _$jscoverage['util/bignum.js'][207]++;
  k += buffer[offset + 3] << 24 >>> 0;
  _$jscoverage['util/bignum.js'][208]++;
  offset += 4;
  _$jscoverage['util/bignum.js'][210]++;
  l = (buffer[offset + 2] & 1) << 16;
  _$jscoverage['util/bignum.js'][211]++;
  l |= buffer[offset + 1] << 8;
  _$jscoverage['util/bignum.js'][212]++;
  l |= buffer[offset];
  _$jscoverage['util/bignum.js'][213]++;
  offset += 2;
  _$jscoverage['util/bignum.js'][215]++;
  var dec = {s: (buffer[offset + 1] & 128)? -1: 1, m: undefined, e: ((((buffer[offset + 1] << 8) | buffer[offset]) & 32766) >> 1) - EXP_BIAS};
  _$jscoverage['util/bignum.js'][222]++;
  if (k === 0 && l === 0) {
    _$jscoverage['util/bignum.js'][223]++;
    if (j === 0) {
      _$jscoverage['util/bignum.js'][224]++;
      dec.m = i;
      _$jscoverage['util/bignum.js'][225]++;
      return dec;
    }
    _$jscoverage['util/bignum.js'][227]++;
    if (j < INT_2_21 || (j === INT_2_21 && i === 0)) {
      _$jscoverage['util/bignum.js'][228]++;
      dec.m = j * INT_2_32 + i;
      _$jscoverage['util/bignum.js'][229]++;
      return dec;
    }
  }
  _$jscoverage['util/bignum.js'][233]++;
  if (i < BASE) {
    _$jscoverage['util/bignum.js'][234]++;
    x0 = i;
    _$jscoverage['util/bignum.js'][235]++;
    x1 = 0;
  }
  else {
    _$jscoverage['util/bignum.js'][237]++;
    x0 = i % BASE;
    _$jscoverage['util/bignum.js'][238]++;
    x1 = Math.floor(i / BASE);
  }
  _$jscoverage['util/bignum.js'][241]++;
  if (j < BASE) {
    _$jscoverage['util/bignum.js'][242]++;
    x0 += j * INT_2_32_0;
    _$jscoverage['util/bignum.js'][243]++;
    x1 += j * INT_2_32_1;
    _$jscoverage['util/bignum.js'][244]++;
    x2 = 0;
  }
  else {
    _$jscoverage['util/bignum.js'][246]++;
    z0 = j % BASE;
    _$jscoverage['util/bignum.js'][247]++;
    z1 = Math.floor(j / BASE);
    _$jscoverage['util/bignum.js'][248]++;
    x0 += z0 * INT_2_32_0;
    _$jscoverage['util/bignum.js'][249]++;
    x1 += z0 * INT_2_32_1 + z1 * INT_2_32_0;
    _$jscoverage['util/bignum.js'][250]++;
    x2 = z1 * INT_2_32_1;
  }
  _$jscoverage['util/bignum.js'][253]++;
  if (k < BASE) {
    _$jscoverage['util/bignum.js'][254]++;
    y0 = k;
    _$jscoverage['util/bignum.js'][255]++;
    y1 = 0;
  }
  else {
    _$jscoverage['util/bignum.js'][257]++;
    y0 = k % BASE;
    _$jscoverage['util/bignum.js'][258]++;
    y1 = Math.floor(k / BASE);
  }
  _$jscoverage['util/bignum.js'][261]++;
  if (l < BASE) {
    _$jscoverage['util/bignum.js'][262]++;
    y0 += l * INT_2_32_0;
    _$jscoverage['util/bignum.js'][263]++;
    y1 += l * INT_2_32_1;
    _$jscoverage['util/bignum.js'][264]++;
    y2 = 0;
  }
  else {
    _$jscoverage['util/bignum.js'][266]++;
    z0 = l % BASE;
    _$jscoverage['util/bignum.js'][267]++;
    z1 = Math.floor(l / BASE);
    _$jscoverage['util/bignum.js'][268]++;
    y0 += z0 * INT_2_32_0;
    _$jscoverage['util/bignum.js'][269]++;
    y1 += z0 * INT_2_32_1 + z1 * INT_2_32_0;
    _$jscoverage['util/bignum.js'][270]++;
    y2 = z1 * INT_2_32_1;
  }
  _$jscoverage['util/bignum.js'][273]++;
  if (y0 >= BASE) {
    _$jscoverage['util/bignum.js'][274]++;
    y1 += Math.floor(y0 / BASE);
    _$jscoverage['util/bignum.js'][275]++;
    y0 %= BASE;
  }
  _$jscoverage['util/bignum.js'][277]++;
  if (y1 >= BASE) {
    _$jscoverage['util/bignum.js'][278]++;
    y2 += Math.floor(y1 / BASE);
    _$jscoverage['util/bignum.js'][279]++;
    y1 %= BASE;
  }
  _$jscoverage['util/bignum.js'][282]++;
  x0 += y0 * INT_2_64_0;
  _$jscoverage['util/bignum.js'][283]++;
  x1 += y0 * INT_2_64_1 + y1 * INT_2_64_0;
  _$jscoverage['util/bignum.js'][284]++;
  x2 += y0 * INT_2_64_2 + y1 * INT_2_64_1 + y2 * INT_2_64_0;
  _$jscoverage['util/bignum.js'][285]++;
  x3 = y1 * INT_2_64_2 + y2 * INT_2_64_1;
  _$jscoverage['util/bignum.js'][286]++;
  x4 = y2 * INT_2_64_2;
  _$jscoverage['util/bignum.js'][288]++;
  if (x0 >= BASE) {
    _$jscoverage['util/bignum.js'][289]++;
    x1 += Math.floor(x0 / BASE);
    _$jscoverage['util/bignum.js'][290]++;
    x0 %= BASE;
  }
  _$jscoverage['util/bignum.js'][292]++;
  if (x1 >= BASE) {
    _$jscoverage['util/bignum.js'][293]++;
    x2 += Math.floor(x1 / BASE);
    _$jscoverage['util/bignum.js'][294]++;
    x1 %= BASE;
  }
  _$jscoverage['util/bignum.js'][296]++;
  if (x2 >= BASE) {
    _$jscoverage['util/bignum.js'][297]++;
    x3 += Math.floor(x2 / BASE);
    _$jscoverage['util/bignum.js'][298]++;
    x2 %= BASE;
  }
  _$jscoverage['util/bignum.js'][300]++;
  if (x3 >= BASE) {
    _$jscoverage['util/bignum.js'][301]++;
    x4 += Math.floor(x3 / BASE);
    _$jscoverage['util/bignum.js'][302]++;
    x3 %= BASE;
  }
  _$jscoverage['util/bignum.js'][305]++;
  if (x4) {
    _$jscoverage['util/bignum.js'][306]++;
    dec.m = "" + x4 + zeropad.lpad14(x3 * BASE + x2) + zeropad.lpad14(x1 * BASE + x0);
    _$jscoverage['util/bignum.js'][308]++;
    return dec;
  }
  _$jscoverage['util/bignum.js'][310]++;
  if (x3) {
    _$jscoverage['util/bignum.js'][311]++;
    dec.m = "" + (x3 * BASE + x2) + zeropad.lpad14(x1 * BASE + x0);
    _$jscoverage['util/bignum.js'][312]++;
    return dec;
  }
  _$jscoverage['util/bignum.js'][314]++;
  if (x2) {
    _$jscoverage['util/bignum.js'][315]++;
    if (x2 < INT_2_53_2 || (x2 === INT_2_53_2 && (x1 < INT_2_53_1 || (x1 === INT_2_53_1 && x0 <= INT_2_53_0)))) {
      _$jscoverage['util/bignum.js'][317]++;
      dec.m = (x2 * BASE + x1) * BASE + x0;
      _$jscoverage['util/bignum.js'][318]++;
      return dec;
    }
    _$jscoverage['util/bignum.js'][320]++;
    dec.m = "" + x2 + zeropad.lpad14(x1 * BASE + x0);
    _$jscoverage['util/bignum.js'][321]++;
    return dec;
  }
  _$jscoverage['util/bignum.js'][323]++;
  dec.m = x1 * BASE + x0;
  _$jscoverage['util/bignum.js'][324]++;
  return dec;
}
_$jscoverage['util/bignum.js'][328]++;
function readDecFloat(buffer, offset) {
  _$jscoverage['util/bignum.js'][329]++;
  var value = readDec128(buffer, offset);
  _$jscoverage['util/bignum.js'][330]++;
  if (value === null) {
    _$jscoverage['util/bignum.js'][331]++;
    return null;
  }
  _$jscoverage['util/bignum.js'][333]++;
  var d, l, e;
  _$jscoverage['util/bignum.js'][334]++;
  d = "" + value.m;
  _$jscoverage['util/bignum.js'][335]++;
  l = d.length;
  _$jscoverage['util/bignum.js'][336]++;
  e = value.e + l - 1;
  _$jscoverage['util/bignum.js'][337]++;
  if (e < 0) {
    _$jscoverage['util/bignum.js'][338]++;
    e = "e" + e;
  }
  else {
    _$jscoverage['util/bignum.js'][340]++;
    e = "e+" + e;
  }
  _$jscoverage['util/bignum.js'][342]++;
  while (d.charCodeAt(l - 1) === 48) {
    _$jscoverage['util/bignum.js'][343]++;
    l -= 1;
}
  _$jscoverage['util/bignum.js'][345]++;
  if (l > 1) {
    _$jscoverage['util/bignum.js'][346]++;
    d = d.substring(0, 1) + "." + d.substring(1, l);
  }
  _$jscoverage['util/bignum.js'][348]++;
  if (value.s < 0) {
    _$jscoverage['util/bignum.js'][349]++;
    return "-" + d + e;
  }
  _$jscoverage['util/bignum.js'][351]++;
  return d + e;
}
_$jscoverage['util/bignum.js'][354]++;
function readDecFixed(buffer, offset, frac) {
  _$jscoverage['util/bignum.js'][355]++;
  var value = readDec128(buffer, offset);
  _$jscoverage['util/bignum.js'][356]++;
  if (value === null) {
    _$jscoverage['util/bignum.js'][357]++;
    return null;
  }
  _$jscoverage['util/bignum.js'][359]++;
  frac = frac || 0;
  _$jscoverage['util/bignum.js'][360]++;
  var d, e, l, k, i, f;
  _$jscoverage['util/bignum.js'][361]++;
  d = "" + value.m;
  _$jscoverage['util/bignum.js'][362]++;
  e = value.e;
  _$jscoverage['util/bignum.js'][363]++;
  if (e < 0) {
    _$jscoverage['util/bignum.js'][364]++;
    l = d.length;
    _$jscoverage['util/bignum.js'][365]++;
    k = l + e;
    _$jscoverage['util/bignum.js'][366]++;
    if (k > 0) {
      _$jscoverage['util/bignum.js'][367]++;
      i = d.substring(0, k);
      _$jscoverage['util/bignum.js'][368]++;
      f = d.substring(k);
    }
    else {
      _$jscoverage['util/bignum.js'][369]++;
      if (k < 0) {
        _$jscoverage['util/bignum.js'][370]++;
        i = "0";
        _$jscoverage['util/bignum.js'][371]++;
        f = zeropad.ZEROS[- k] + d;
      }
      else {
        _$jscoverage['util/bignum.js'][373]++;
        i = "0";
        _$jscoverage['util/bignum.js'][374]++;
        f = d;
      }
    }
  }
  else {
    _$jscoverage['util/bignum.js'][376]++;
    if (e > 0) {
      _$jscoverage['util/bignum.js'][377]++;
      i = d + zeropad.ZEROS[e];
      _$jscoverage['util/bignum.js'][378]++;
      f = "";
    }
    else {
      _$jscoverage['util/bignum.js'][380]++;
      i = d;
      _$jscoverage['util/bignum.js'][381]++;
      f = "";
    }
  }
  _$jscoverage['util/bignum.js'][383]++;
  if (value.s < 0) {
    _$jscoverage['util/bignum.js'][384]++;
    i = "-" + i;
  }
  _$jscoverage['util/bignum.js'][386]++;
  if (frac === 0) {
    _$jscoverage['util/bignum.js'][387]++;
    return i;
  }
  _$jscoverage['util/bignum.js'][389]++;
  l = f.length;
  _$jscoverage['util/bignum.js'][390]++;
  if (l > frac) {
    _$jscoverage['util/bignum.js'][391]++;
    f = f.substring(0, frac);
  }
  else {
    _$jscoverage['util/bignum.js'][392]++;
    if (l < frac) {
      _$jscoverage['util/bignum.js'][393]++;
      f += zeropad.ZEROS[frac - l];
    }
  }
  _$jscoverage['util/bignum.js'][395]++;
  return i + "." + f;
}
_$jscoverage['util/bignum.js'][399]++;
function _writeInt64(buffer, value, offset, unsigned) {
  _$jscoverage['util/bignum.js'][401]++;
  var l, x, a, b, c, x0, x1, y0, y1, z0, z1, z2, z3;
  _$jscoverage['util/bignum.js'][403]++;
  var negative = false;
  _$jscoverage['util/bignum.js'][405]++;
  if (typeof value === "string") {
    _$jscoverage['util/bignum.js'][407]++;
    l = value.length;
    _$jscoverage['util/bignum.js'][409]++;
    if (l > 15) {
      _$jscoverage['util/bignum.js'][411]++;
      a = + value.substring(l - 7);
      _$jscoverage['util/bignum.js'][412]++;
      b = + value.substring(l - 14, l - 7);
      _$jscoverage['util/bignum.js'][413]++;
      c = + value.substring(0, l - 14);
      _$jscoverage['util/bignum.js'][415]++;
      if (! unsigned && c < 0) {
        _$jscoverage['util/bignum.js'][416]++;
        c *= -1;
        _$jscoverage['util/bignum.js'][417]++;
        negative = true;
      }
      _$jscoverage['util/bignum.js'][420]++;
      z0 = a % INT_2_16;
      _$jscoverage['util/bignum.js'][421]++;
      z1 = Math.floor(a / INT_2_16);
      _$jscoverage['util/bignum.js'][422]++;
      z2 = z3 = 0;
      _$jscoverage['util/bignum.js'][424]++;
      y0 = b % INT_2_16;
      _$jscoverage['util/bignum.js'][425]++;
      y1 = Math.floor(b / INT_2_16);
      _$jscoverage['util/bignum.js'][427]++;
      z0 += y0 * BIN_10_7_0;
      _$jscoverage['util/bignum.js'][428]++;
      z1 += y0 * BIN_10_7_1 + y1 * BIN_10_7_0;
      _$jscoverage['util/bignum.js'][429]++;
      z2 += y1 * BIN_10_7_1;
      _$jscoverage['util/bignum.js'][431]++;
      y0 = c % INT_2_16;
      _$jscoverage['util/bignum.js'][432]++;
      y1 = Math.floor(c / INT_2_16);
      _$jscoverage['util/bignum.js'][434]++;
      z0 += y0 * BIN_10_14_0;
      _$jscoverage['util/bignum.js'][435]++;
      z1 += y0 * BIN_10_14_1 + y1 * BIN_10_14_0;
      _$jscoverage['util/bignum.js'][436]++;
      z2 += y0 * BIN_10_14_2 + y1 * BIN_10_14_1;
      _$jscoverage['util/bignum.js'][437]++;
      z3 += y1 * BIN_10_14_2;
      _$jscoverage['util/bignum.js'][439]++;
      if (z0 >= INT_2_16) {
        _$jscoverage['util/bignum.js'][440]++;
        z1 += Math.floor(z0 / INT_2_16);
        _$jscoverage['util/bignum.js'][441]++;
        z0 %= INT_2_16;
      }
      _$jscoverage['util/bignum.js'][443]++;
      if (z1 >= INT_2_16) {
        _$jscoverage['util/bignum.js'][444]++;
        z2 += Math.floor(z1 / INT_2_16);
        _$jscoverage['util/bignum.js'][445]++;
        z1 %= INT_2_16;
      }
      _$jscoverage['util/bignum.js'][447]++;
      if (z2 >= INT_2_16) {
        _$jscoverage['util/bignum.js'][448]++;
        z3 += Math.floor(z2 / INT_2_16);
        _$jscoverage['util/bignum.js'][449]++;
        z2 %= INT_2_16;
      }
      _$jscoverage['util/bignum.js'][452]++;
      x0 = z1 * INT_2_16 + z0;
      _$jscoverage['util/bignum.js'][453]++;
      x1 = z3 * INT_2_16 + z2;
    }
    else {
      _$jscoverage['util/bignum.js'][456]++;
      x = + value;
      _$jscoverage['util/bignum.js'][458]++;
      if (! unsigned && x < 0) {
        _$jscoverage['util/bignum.js'][459]++;
        x *= -1;
        _$jscoverage['util/bignum.js'][460]++;
        negative = true;
      }
      _$jscoverage['util/bignum.js'][463]++;
      x0 = x % INT_2_32;
      _$jscoverage['util/bignum.js'][464]++;
      x1 = Math.floor(x / INT_2_32);
    }
  }
  else {
    _$jscoverage['util/bignum.js'][467]++;
    x = value;
    _$jscoverage['util/bignum.js'][469]++;
    if (! unsigned && x < 0) {
      _$jscoverage['util/bignum.js'][470]++;
      x *= -1;
      _$jscoverage['util/bignum.js'][471]++;
      negative = true;
    }
    _$jscoverage['util/bignum.js'][474]++;
    x0 = x % INT_2_32;
    _$jscoverage['util/bignum.js'][475]++;
    x1 = Math.floor(x / INT_2_32);
  }
  _$jscoverage['util/bignum.js'][478]++;
  if (negative) {
    _$jscoverage['util/bignum.js'][479]++;
    x1 = 4294967295 - x1;
    _$jscoverage['util/bignum.js'][480]++;
    if (x0 === 0) {
      _$jscoverage['util/bignum.js'][481]++;
      x1 += 1;
    }
    else {
      _$jscoverage['util/bignum.js'][483]++;
      x0 = 4294967295 - x0 + 1;
    }
  }
  _$jscoverage['util/bignum.js'][487]++;
  buffer[offset + 3] = (x0 >>> 24) & 255;
  _$jscoverage['util/bignum.js'][488]++;
  buffer[offset + 2] = (x0 >>> 16) & 255;
  _$jscoverage['util/bignum.js'][489]++;
  buffer[offset + 1] = (x0 >>> 8) & 255;
  _$jscoverage['util/bignum.js'][490]++;
  buffer[offset] = x0 & 255;
  _$jscoverage['util/bignum.js'][491]++;
  offset += 4;
  _$jscoverage['util/bignum.js'][493]++;
  buffer[offset + 3] = (x1 >>> 24) & 255;
  _$jscoverage['util/bignum.js'][494]++;
  buffer[offset + 2] = (x1 >>> 16) & 255;
  _$jscoverage['util/bignum.js'][495]++;
  buffer[offset + 1] = (x1 >>> 8) & 255;
  _$jscoverage['util/bignum.js'][496]++;
  buffer[offset] = x1 & 255;
}
_$jscoverage['util/bignum.js'][499]++;
function readUInt128(buffer, offset) {
  _$jscoverage['util/bignum.js'][501]++;
  var i, j, k, l, z0, z1, y0, y1, y2, x0, x1, x2, x3, x4;
  _$jscoverage['util/bignum.js'][503]++;
  offset = offset || 0;
  _$jscoverage['util/bignum.js'][505]++;
  i = buffer[offset + 2] << 16;
  _$jscoverage['util/bignum.js'][506]++;
  i |= buffer[offset + 1] << 8;
  _$jscoverage['util/bignum.js'][507]++;
  i |= buffer[offset];
  _$jscoverage['util/bignum.js'][508]++;
  i += buffer[offset + 3] << 24 >>> 0;
  _$jscoverage['util/bignum.js'][509]++;
  offset += 4;
  _$jscoverage['util/bignum.js'][511]++;
  j = buffer[offset + 2] << 16;
  _$jscoverage['util/bignum.js'][512]++;
  j |= buffer[offset + 1] << 8;
  _$jscoverage['util/bignum.js'][513]++;
  j |= buffer[offset];
  _$jscoverage['util/bignum.js'][514]++;
  j += buffer[offset + 3] << 24 >>> 0;
  _$jscoverage['util/bignum.js'][515]++;
  offset += 4;
  _$jscoverage['util/bignum.js'][517]++;
  k = buffer[offset + 2] << 16;
  _$jscoverage['util/bignum.js'][518]++;
  k |= buffer[offset + 1] << 8;
  _$jscoverage['util/bignum.js'][519]++;
  k |= buffer[offset];
  _$jscoverage['util/bignum.js'][520]++;
  k += buffer[offset + 3] << 24 >>> 0;
  _$jscoverage['util/bignum.js'][521]++;
  offset += 4;
  _$jscoverage['util/bignum.js'][523]++;
  l = buffer[offset + 2] << 16;
  _$jscoverage['util/bignum.js'][524]++;
  l |= buffer[offset + 1] << 8;
  _$jscoverage['util/bignum.js'][525]++;
  l |= buffer[offset];
  _$jscoverage['util/bignum.js'][526]++;
  l += buffer[offset + 3] << 24 >>> 0;
  _$jscoverage['util/bignum.js'][528]++;
  if (k === 0 && l === 0) {
    _$jscoverage['util/bignum.js'][529]++;
    if (j === 0) {
      _$jscoverage['util/bignum.js'][530]++;
      return i;
    }
    _$jscoverage['util/bignum.js'][532]++;
    if (j < INT_2_21 || (j === INT_2_21 && i === 0)) {
      _$jscoverage['util/bignum.js'][533]++;
      return j * INT_2_32 + i;
    }
  }
  _$jscoverage['util/bignum.js'][537]++;
  if (i < BASE) {
    _$jscoverage['util/bignum.js'][538]++;
    x0 = i;
    _$jscoverage['util/bignum.js'][539]++;
    x1 = 0;
  }
  else {
    _$jscoverage['util/bignum.js'][541]++;
    x0 = i % BASE;
    _$jscoverage['util/bignum.js'][542]++;
    x1 = Math.floor(i / BASE);
  }
  _$jscoverage['util/bignum.js'][545]++;
  if (j < BASE) {
    _$jscoverage['util/bignum.js'][546]++;
    x0 += j * INT_2_32_0;
    _$jscoverage['util/bignum.js'][547]++;
    x1 += j * INT_2_32_1;
    _$jscoverage['util/bignum.js'][548]++;
    x2 = 0;
  }
  else {
    _$jscoverage['util/bignum.js'][550]++;
    z0 = j % BASE;
    _$jscoverage['util/bignum.js'][551]++;
    z1 = Math.floor(j / BASE);
    _$jscoverage['util/bignum.js'][552]++;
    x0 += z0 * INT_2_32_0;
    _$jscoverage['util/bignum.js'][553]++;
    x1 += z0 * INT_2_32_1 + z1 * INT_2_32_0;
    _$jscoverage['util/bignum.js'][554]++;
    x2 = z1 * INT_2_32_1;
  }
  _$jscoverage['util/bignum.js'][557]++;
  if (k < BASE) {
    _$jscoverage['util/bignum.js'][558]++;
    y0 = k;
    _$jscoverage['util/bignum.js'][559]++;
    y1 = 0;
  }
  else {
    _$jscoverage['util/bignum.js'][561]++;
    y0 = k % BASE;
    _$jscoverage['util/bignum.js'][562]++;
    y1 = Math.floor(k / BASE);
  }
  _$jscoverage['util/bignum.js'][565]++;
  if (l < BASE) {
    _$jscoverage['util/bignum.js'][566]++;
    y0 += l * INT_2_32_0;
    _$jscoverage['util/bignum.js'][567]++;
    y1 += l * INT_2_32_1;
    _$jscoverage['util/bignum.js'][568]++;
    y2 = 0;
  }
  else {
    _$jscoverage['util/bignum.js'][570]++;
    z0 = l % BASE;
    _$jscoverage['util/bignum.js'][571]++;
    z1 = Math.floor(l / BASE);
    _$jscoverage['util/bignum.js'][572]++;
    y0 += z0 * INT_2_32_0;
    _$jscoverage['util/bignum.js'][573]++;
    y1 += z0 * INT_2_32_1 + z1 * INT_2_32_0;
    _$jscoverage['util/bignum.js'][574]++;
    y2 = z1 * INT_2_32_1;
  }
  _$jscoverage['util/bignum.js'][577]++;
  if (y0 >= BASE) {
    _$jscoverage['util/bignum.js'][578]++;
    y1 += Math.floor(y0 / BASE);
    _$jscoverage['util/bignum.js'][579]++;
    y0 %= BASE;
  }
  _$jscoverage['util/bignum.js'][581]++;
  if (y1 >= BASE) {
    _$jscoverage['util/bignum.js'][582]++;
    y2 += Math.floor(y1 / BASE);
    _$jscoverage['util/bignum.js'][583]++;
    y1 %= BASE;
  }
  _$jscoverage['util/bignum.js'][586]++;
  x0 += y0 * INT_2_64_0;
  _$jscoverage['util/bignum.js'][587]++;
  x1 += y0 * INT_2_64_1 + y1 * INT_2_64_0;
  _$jscoverage['util/bignum.js'][588]++;
  x2 += y0 * INT_2_64_2 + y1 * INT_2_64_1 + y2 * INT_2_64_0;
  _$jscoverage['util/bignum.js'][589]++;
  x3 = y1 * INT_2_64_2 + y2 * INT_2_64_1;
  _$jscoverage['util/bignum.js'][590]++;
  x4 = y2 * INT_2_64_2;
  _$jscoverage['util/bignum.js'][592]++;
  if (x0 >= BASE) {
    _$jscoverage['util/bignum.js'][593]++;
    x1 += Math.floor(x0 / BASE);
    _$jscoverage['util/bignum.js'][594]++;
    x0 %= BASE;
  }
  _$jscoverage['util/bignum.js'][596]++;
  if (x1 >= BASE) {
    _$jscoverage['util/bignum.js'][597]++;
    x2 += Math.floor(x1 / BASE);
    _$jscoverage['util/bignum.js'][598]++;
    x1 %= BASE;
  }
  _$jscoverage['util/bignum.js'][600]++;
  if (x2 >= BASE) {
    _$jscoverage['util/bignum.js'][601]++;
    x3 += Math.floor(x2 / BASE);
    _$jscoverage['util/bignum.js'][602]++;
    x2 %= BASE;
  }
  _$jscoverage['util/bignum.js'][604]++;
  if (x3 >= BASE) {
    _$jscoverage['util/bignum.js'][605]++;
    x4 += Math.floor(x3 / BASE);
    _$jscoverage['util/bignum.js'][606]++;
    x3 %= BASE;
  }
  _$jscoverage['util/bignum.js'][609]++;
  if (x4) {
    _$jscoverage['util/bignum.js'][610]++;
    return "" + x4 + zeropad.lpad14(x3 * BASE + x2) + zeropad.lpad14(x1 * BASE + x0);
  }
  _$jscoverage['util/bignum.js'][613]++;
  if (x3) {
    _$jscoverage['util/bignum.js'][614]++;
    return "" + (x3 * BASE + x2) + zeropad.lpad14(x1 * BASE + x0);
  }
  _$jscoverage['util/bignum.js'][616]++;
  if (x2) {
    _$jscoverage['util/bignum.js'][617]++;
    if (x2 < INT_2_53_2 || (x2 === INT_2_53_2 && (x1 < INT_2_53_1 || (x1 === INT_2_53_1 && x0 <= INT_2_53_0)))) {
      _$jscoverage['util/bignum.js'][619]++;
      return (x2 * BASE + x1) * BASE + x0;
    }
    _$jscoverage['util/bignum.js'][621]++;
    return "" + x2 + zeropad.lpad14(x1 * BASE + x0);
  }
  _$jscoverage['util/bignum.js'][624]++;
  return x1 * BASE + x0;
}
_$jscoverage['util/bignum.js'][628]++;
function writeUInt128(buffer, value, offset) {
  _$jscoverage['util/bignum.js'][630]++;
  var l, a, b, c, d, e, f, x0, x1, x2, x3, y0, y1, z0, z1, z2, z3, z4, z5, z6, z7;
  _$jscoverage['util/bignum.js'][633]++;
  l = value.length;
  _$jscoverage['util/bignum.js'][635]++;
  a = + value.substring(l - 7);
  _$jscoverage['util/bignum.js'][636]++;
  if (l > 7) {
    _$jscoverage['util/bignum.js'][637]++;
    b = + value.substring(l - 14, l - 7);
  }
  else {
    _$jscoverage['util/bignum.js'][639]++;
    b = 0;
  }
  _$jscoverage['util/bignum.js'][641]++;
  if (l > 14) {
    _$jscoverage['util/bignum.js'][642]++;
    c = + value.substring(l - 21, l - 14);
  }
  else {
    _$jscoverage['util/bignum.js'][644]++;
    c = 0;
  }
  _$jscoverage['util/bignum.js'][646]++;
  if (l > 21) {
    _$jscoverage['util/bignum.js'][647]++;
    d = + value.substring(l - 28, l - 21);
  }
  else {
    _$jscoverage['util/bignum.js'][649]++;
    d = 0;
  }
  _$jscoverage['util/bignum.js'][651]++;
  if (l > 28) {
    _$jscoverage['util/bignum.js'][652]++;
    e = + value.substring(l - 35, l - 28);
  }
  else {
    _$jscoverage['util/bignum.js'][654]++;
    e = 0;
  }
  _$jscoverage['util/bignum.js'][656]++;
  if (l > 35) {
    _$jscoverage['util/bignum.js'][657]++;
    f = + value.substring(0, l - 35);
  }
  else {
    _$jscoverage['util/bignum.js'][659]++;
    f = 0;
  }
  _$jscoverage['util/bignum.js'][662]++;
  z0 = z1 = z2 = z3 = z4 = z5 = z6 = z7 = 0;
  _$jscoverage['util/bignum.js'][665]++;
  if (a) {
    _$jscoverage['util/bignum.js'][666]++;
    z0 = a % INT_2_16;
    _$jscoverage['util/bignum.js'][667]++;
    z1 = Math.floor(a / INT_2_16);
  }
  _$jscoverage['util/bignum.js'][671]++;
  if (b) {
    _$jscoverage['util/bignum.js'][672]++;
    y0 = b % INT_2_16;
    _$jscoverage['util/bignum.js'][673]++;
    y1 = Math.floor(b / INT_2_16);
    _$jscoverage['util/bignum.js'][674]++;
    z0 += y0 * BIN_10_7_0;
    _$jscoverage['util/bignum.js'][675]++;
    z1 += y0 * BIN_10_7_1 + y1 * BIN_10_7_0;
    _$jscoverage['util/bignum.js'][676]++;
    z2 += y1 * BIN_10_7_1;
  }
  _$jscoverage['util/bignum.js'][680]++;
  if (c) {
    _$jscoverage['util/bignum.js'][681]++;
    y0 = c % INT_2_16;
    _$jscoverage['util/bignum.js'][682]++;
    y1 = Math.floor(c / INT_2_16);
    _$jscoverage['util/bignum.js'][683]++;
    z0 += y0 * BIN_10_14_0;
    _$jscoverage['util/bignum.js'][684]++;
    z1 += y0 * BIN_10_14_1 + y1 * BIN_10_14_0;
    _$jscoverage['util/bignum.js'][685]++;
    z2 += y0 * BIN_10_14_2 + y1 * BIN_10_14_1;
    _$jscoverage['util/bignum.js'][686]++;
    z3 += y1 * BIN_10_14_2;
  }
  _$jscoverage['util/bignum.js'][690]++;
  if (d) {
    _$jscoverage['util/bignum.js'][691]++;
    y0 = d % INT_2_16;
    _$jscoverage['util/bignum.js'][692]++;
    y1 = Math.floor(d / INT_2_16);
    _$jscoverage['util/bignum.js'][693]++;
    z0 += y0 * BIN_10_21_0;
    _$jscoverage['util/bignum.js'][694]++;
    z1 += y0 * BIN_10_21_1 + y1 * BIN_10_21_0;
    _$jscoverage['util/bignum.js'][695]++;
    z2 += y0 * BIN_10_21_2 + y1 * BIN_10_21_1;
    _$jscoverage['util/bignum.js'][696]++;
    z3 += y0 * BIN_10_21_3 + y1 * BIN_10_21_2;
    _$jscoverage['util/bignum.js'][697]++;
    z4 += y0 * BIN_10_21_4 + y1 * BIN_10_21_3;
    _$jscoverage['util/bignum.js'][698]++;
    z5 += y1 * BIN_10_21_4;
  }
  _$jscoverage['util/bignum.js'][702]++;
  if (e) {
    _$jscoverage['util/bignum.js'][703]++;
    y0 = e % INT_2_16;
    _$jscoverage['util/bignum.js'][704]++;
    y1 = Math.floor(e / INT_2_16);
    _$jscoverage['util/bignum.js'][705]++;
    z0 += y0 * BIN_10_28_0;
    _$jscoverage['util/bignum.js'][706]++;
    z1 += y0 * BIN_10_28_1 + y1 * BIN_10_28_0;
    _$jscoverage['util/bignum.js'][707]++;
    z2 += y0 * BIN_10_28_2 + y1 * BIN_10_28_1;
    _$jscoverage['util/bignum.js'][708]++;
    z3 += y0 * BIN_10_28_3 + y1 * BIN_10_28_2;
    _$jscoverage['util/bignum.js'][709]++;
    z4 += y0 * BIN_10_28_4 + y1 * BIN_10_28_3;
    _$jscoverage['util/bignum.js'][710]++;
    z5 += y0 * BIN_10_28_5 + y1 * BIN_10_28_4;
    _$jscoverage['util/bignum.js'][711]++;
    z6 += y1 * BIN_10_28_5;
  }
  _$jscoverage['util/bignum.js'][715]++;
  if (f) {
    _$jscoverage['util/bignum.js'][716]++;
    y0 = f % INT_2_16;
    _$jscoverage['util/bignum.js'][717]++;
    z0 += y0 * BIN_10_35_0;
    _$jscoverage['util/bignum.js'][718]++;
    z1 += y0 * BIN_10_35_1;
    _$jscoverage['util/bignum.js'][719]++;
    z2 += y0 * BIN_10_35_2;
    _$jscoverage['util/bignum.js'][720]++;
    z3 += y0 * BIN_10_35_3;
    _$jscoverage['util/bignum.js'][721]++;
    z4 += y0 * BIN_10_35_4;
    _$jscoverage['util/bignum.js'][722]++;
    z5 += y0 * BIN_10_35_5;
    _$jscoverage['util/bignum.js'][723]++;
    z6 += y0 * BIN_10_35_6;
    _$jscoverage['util/bignum.js'][724]++;
    z7 += y0 * BIN_10_35_7;
  }
  _$jscoverage['util/bignum.js'][727]++;
  if (z0 >= INT_2_16) {
    _$jscoverage['util/bignum.js'][728]++;
    z1 += Math.floor(z0 / INT_2_16);
    _$jscoverage['util/bignum.js'][729]++;
    z0 %= INT_2_16;
  }
  _$jscoverage['util/bignum.js'][731]++;
  if (z1 >= INT_2_16) {
    _$jscoverage['util/bignum.js'][732]++;
    z2 += Math.floor(z1 / INT_2_16);
    _$jscoverage['util/bignum.js'][733]++;
    z1 %= INT_2_16;
  }
  _$jscoverage['util/bignum.js'][735]++;
  if (z2 >= INT_2_16) {
    _$jscoverage['util/bignum.js'][736]++;
    z3 += Math.floor(z2 / INT_2_16);
    _$jscoverage['util/bignum.js'][737]++;
    z2 %= INT_2_16;
  }
  _$jscoverage['util/bignum.js'][739]++;
  if (z3 >= INT_2_16) {
    _$jscoverage['util/bignum.js'][740]++;
    z4 += Math.floor(z3 / INT_2_16);
    _$jscoverage['util/bignum.js'][741]++;
    z3 %= INT_2_16;
  }
  _$jscoverage['util/bignum.js'][743]++;
  if (z4 >= INT_2_16) {
    _$jscoverage['util/bignum.js'][744]++;
    z5 += Math.floor(z4 / INT_2_16);
    _$jscoverage['util/bignum.js'][745]++;
    z4 %= INT_2_16;
  }
  _$jscoverage['util/bignum.js'][747]++;
  if (z5 >= INT_2_16) {
    _$jscoverage['util/bignum.js'][748]++;
    z6 += Math.floor(z5 / INT_2_16);
    _$jscoverage['util/bignum.js'][749]++;
    z5 %= INT_2_16;
  }
  _$jscoverage['util/bignum.js'][751]++;
  if (z6 >= INT_2_16) {
    _$jscoverage['util/bignum.js'][752]++;
    z7 += Math.floor(z6 / INT_2_16);
    _$jscoverage['util/bignum.js'][753]++;
    z6 %= INT_2_16;
  }
  _$jscoverage['util/bignum.js'][756]++;
  x0 = z1 * INT_2_16 + z0;
  _$jscoverage['util/bignum.js'][757]++;
  x1 = z3 * INT_2_16 + z2;
  _$jscoverage['util/bignum.js'][758]++;
  x2 = z5 * INT_2_16 + z4;
  _$jscoverage['util/bignum.js'][759]++;
  x3 = z7 * INT_2_16 + z6;
  _$jscoverage['util/bignum.js'][761]++;
  buffer[offset + 3] = (x0 >>> 24) & 255;
  _$jscoverage['util/bignum.js'][762]++;
  buffer[offset + 2] = (x0 >>> 16) & 255;
  _$jscoverage['util/bignum.js'][763]++;
  buffer[offset + 1] = (x0 >>> 8) & 255;
  _$jscoverage['util/bignum.js'][764]++;
  buffer[offset] = x0 & 255;
  _$jscoverage['util/bignum.js'][765]++;
  offset += 4;
  _$jscoverage['util/bignum.js'][767]++;
  buffer[offset + 3] = (x1 >>> 24) & 255;
  _$jscoverage['util/bignum.js'][768]++;
  buffer[offset + 2] = (x1 >>> 16) & 255;
  _$jscoverage['util/bignum.js'][769]++;
  buffer[offset + 1] = (x1 >>> 8) & 255;
  _$jscoverage['util/bignum.js'][770]++;
  buffer[offset] = x1 & 255;
  _$jscoverage['util/bignum.js'][771]++;
  offset += 4;
  _$jscoverage['util/bignum.js'][773]++;
  buffer[offset + 3] = (x2 >>> 24) & 255;
  _$jscoverage['util/bignum.js'][774]++;
  buffer[offset + 2] = (x2 >>> 16) & 255;
  _$jscoverage['util/bignum.js'][775]++;
  buffer[offset + 1] = (x2 >>> 8) & 255;
  _$jscoverage['util/bignum.js'][776]++;
  buffer[offset] = x2 & 255;
  _$jscoverage['util/bignum.js'][777]++;
  offset += 4;
  _$jscoverage['util/bignum.js'][779]++;
  buffer[offset + 3] = (x3 >>> 24) & 255;
  _$jscoverage['util/bignum.js'][780]++;
  buffer[offset + 2] = (x3 >>> 16) & 255;
  _$jscoverage['util/bignum.js'][781]++;
  buffer[offset + 1] = (x3 >>> 8) & 255;
  _$jscoverage['util/bignum.js'][782]++;
  buffer[offset] = x3 & 255;
}
_$jscoverage['util/bignum.js'][785]++;
function readInt64(buffer, offset) {
  _$jscoverage['util/bignum.js'][786]++;
  return _readInt64(buffer, offset || 0, false);
}
_$jscoverage['util/bignum.js'][789]++;
function readUInt64(buffer, offset) {
  _$jscoverage['util/bignum.js'][790]++;
  return _readInt64(buffer, offset || 0, true);
}
_$jscoverage['util/bignum.js'][793]++;
function writeInt64(buffer, value, offset) {
  _$jscoverage['util/bignum.js'][794]++;
  _writeInt64(buffer, value, offset || 0, false);
}
_$jscoverage['util/bignum.js'][797]++;
function writeUInt64(buffer, value, offset) {
  _$jscoverage['util/bignum.js'][798]++;
  _writeInt64(buffer, value, offset || 0, true);
}
_$jscoverage['util/bignum.js'][801]++;
exports.readInt64LE = readInt64;
_$jscoverage['util/bignum.js'][802]++;
exports.readUInt64LE = readUInt64;
_$jscoverage['util/bignum.js'][803]++;
exports.writeInt64LE = writeInt64;
_$jscoverage['util/bignum.js'][804]++;
exports.writeUInt64LE = writeUInt64;
_$jscoverage['util/bignum.js'][805]++;
exports.readUInt128LE = readUInt128;
_$jscoverage['util/bignum.js'][806]++;
exports.writeUInt128LE = writeUInt128;
_$jscoverage['util/bignum.js'][807]++;
exports.readDec128 = readDec128;
_$jscoverage['util/bignum.js'][808]++;
exports.writeDec128 = writeDec128;
_$jscoverage['util/bignum.js'][809]++;
exports.readDecFloat = readDecFloat;
_$jscoverage['util/bignum.js'][810]++;
exports.readDecFixed = readDecFixed;
_$jscoverage['util/bignum.js'].source = ["// Copyright 2013 SAP AG.","//","// Licensed under the Apache License, Version 2.0 (the \"License\");","// you may not use this file except in compliance with the License.","// You may obtain a copy of the License at","//","// http: //www.apache.org/licenses/LICENSE-2.0","//","// Unless required by applicable law or agreed to in writing,","// software distributed under the License is distributed on an","// \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,","// either express or implied. See the License for the specific","// language governing permissions and limitations under the License.","'use strict';","/* jshint bitwise:false */","","var zeropad = require('./zeropad');","","var BASE = Math.pow(10, 7);","","var EXP_BIAS = 6176;","","var INT_2_16 = Math.pow(2, 16);","var INT_2_21 = Math.pow(2, 21);","var INT_2_32 = Math.pow(2, 32);","","/* 2^32 = 429 4967296 */","var INT_2_32_0 = 4967296;","var INT_2_32_1 = 429;","","/* 2^53 = 90 0719925 4740992 */","var INT_2_53_0 = 4740992;","var INT_2_53_1 = 719925;","var INT_2_53_2 = 90;","","/* 2^64 = 184467 4407370 9551616 */","var INT_2_64_0 = 9551616;","var INT_2_64_1 = 4407370;","var INT_2_64_2 = 184467;","","/* 10^7 for base 2^16 */","var BIN_10_7_0 = 38528;","var BIN_10_7_1 = 152;","","/* 10^14 for base 2^16 */","var BIN_10_14_0 = 16384;","var BIN_10_14_1 = 4218;","var BIN_10_14_2 = 23283;","","/* 10^21 for base 2^16 */","var BIN_10_21_0 = 0;","var BIN_10_21_1 = 56992;","var BIN_10_21_2 = 44485;","var BIN_10_21_3 = 13769;","var BIN_10_21_4 = 54;","","/* 10^28 for base 2^16 */","var BIN_10_28_0 = 0;","var BIN_10_28_1 = 4096;","var BIN_10_28_2 = 609;","var BIN_10_28_3 = 15909;","var BIN_10_28_4 = 52830;","var BIN_10_28_5 = 8271;","","/* 10^35 for base 2^16 */","var BIN_10_35_0 = 0;","var BIN_10_35_1 = 0;","var BIN_10_35_2 = 36840;","var BIN_10_35_3 = 11143;","var BIN_10_35_4 = 19842;","var BIN_10_35_5 = 29383;","var BIN_10_35_6 = 16993;","var BIN_10_35_7 = 19;","","function _readInt64(buffer, offset, unsigned) {","","  var x, y, s, y0, y1, x0, x1, x2;","","  x = buffer[offset + 2] &lt;&lt; 16;","  x |= buffer[offset + 1] &lt;&lt; 8;","  x |= buffer[offset];","  x += buffer[offset + 3] &lt;&lt; 24 &gt;&gt;&gt; 0;","  offset += 4;","","  y = buffer[offset + 2] &lt;&lt; 16;","  y |= buffer[offset + 1] &lt;&lt; 8;","  y |= buffer[offset];","  y += buffer[offset + 3] &lt;&lt; 24 &gt;&gt;&gt; 0;","","  if (!unsigned &amp;&amp; (y &amp; 0x80000000)) {","    y = 0xffffffff - y;","    if (x === 0) {","      y += 1;","    } else {","      x = 0xffffffff - x + 1;","    }","    s = -1;","  } else {","    s = 1;","  }","","  if (y === 0) {","    return s * x;","  }","  if (y &lt; INT_2_21 || (y === INT_2_21 &amp;&amp; x === 0)) {","    return s * (y * INT_2_32 + x);","  }","","  if (x &lt; BASE) {","    x0 = x % BASE;","    x1 = 0;","  } else {","    x0 = x % BASE;","    x1 = Math.floor(x / BASE);","  }","","  if (y &lt; BASE) {","    y0 = y % BASE;","    y1 = 0;","  } else {","    y0 = y % BASE;","    y1 = Math.floor(y / BASE);","  }","","  x0 += y0 * INT_2_32_0;","  x1 += y0 * INT_2_32_1 + y1 * INT_2_32_0;","  x2 = y1 * INT_2_32_1;","","  if (x0 &gt;= BASE) {","    x1 += Math.floor(x0 / BASE);","    x0 %= BASE;","  }","  if (x1 &gt;= BASE) {","    x2 += Math.floor(x1 / BASE);","    x1 %= BASE;","  }","","  if (s === 1) {","    return '' + x2 + zeropad.lpad14(x1 * BASE + x0);","  }","  return '-' + x2 + zeropad.lpad14(x1 * BASE + x0);","","}","","function isUInt64(value) {","  if (typeof value === 'number') {","    return true;","  }","  if (value.length &lt; 20) {","    return true;","  }","  if (value.length === 20 &amp;&amp; value &lt;= '18446744073709551616') {","    return true;","  }","  return false;","}","","function writeDec128(buffer, value, offset) {","  offset = offset || 0;","  buffer.fill(0x00, offset, offset + 16);","","  if (isUInt64(value.m)) {","    writeUInt64(buffer, value.m, offset);","  } else {","    writeUInt128(buffer, value.m, offset);","  }","","  var e = EXP_BIAS + value.e;","","  var e0 = e &lt;&lt; 1;","  e0 &amp;= 0xfe;","  e0 |= buffer[offset + 14] &amp; 0x01;","  buffer[offset + 14] = e0;","","  var e1 = e &gt;&gt; 7;","  if (value.s &lt; 0) {","    e1 |= 0x80;","  }","  buffer[offset + 15] = e1;","}","","function readDec128(buffer, offset) {","","  var i, j, k, l, z0, z1, y0, y1, y2, x0, x1, x2, x3, x4;","","  offset = offset || 0;","","  if ((buffer[offset + 15] &amp; 0x70) === 0x70) {","    return null;","  }","","  i = buffer[offset + 2] &lt;&lt; 16;","  i |= buffer[offset + 1] &lt;&lt; 8;","  i |= buffer[offset];","  i += buffer[offset + 3] &lt;&lt; 24 &gt;&gt;&gt; 0;","  offset += 4;","","  j = buffer[offset + 2] &lt;&lt; 16;","  j |= buffer[offset + 1] &lt;&lt; 8;","  j |= buffer[offset];","  j += buffer[offset + 3] &lt;&lt; 24 &gt;&gt;&gt; 0;","  offset += 4;","","  k = buffer[offset + 2] &lt;&lt; 16;","  k |= buffer[offset + 1] &lt;&lt; 8;","  k |= buffer[offset];","  k += buffer[offset + 3] &lt;&lt; 24 &gt;&gt;&gt; 0;","  offset += 4;","","  l = (buffer[offset + 2] &amp; 0x01) &lt;&lt; 16;","  l |= buffer[offset + 1] &lt;&lt; 8;","  l |= buffer[offset];","  offset += 2;","","  var dec = {","    s: (buffer[offset + 1] &amp; 0x80) ? -1 : 1,","    m: undefined,","    e: ((((buffer[offset + 1] &lt;&lt; 8) | buffer[offset]) &amp; 0x7ffe) &gt;&gt; 1) -","      EXP_BIAS","  };","","  if (k === 0 &amp;&amp; l === 0) {","    if (j === 0) {","      dec.m = i;","      return dec;","    }","    if (j &lt; INT_2_21 || (j === INT_2_21 &amp;&amp; i === 0)) {","      dec.m = j * INT_2_32 + i;","      return dec;","    }","  }","","  if (i &lt; BASE) {","    x0 = i;","    x1 = 0;","  } else {","    x0 = i % BASE;","    x1 = Math.floor(i / BASE);","  }","","  if (j &lt; BASE) {","    x0 += j * INT_2_32_0;","    x1 += j * INT_2_32_1;","    x2 = 0;","  } else {","    z0 = j % BASE;","    z1 = Math.floor(j / BASE);","    x0 += z0 * INT_2_32_0;","    x1 += z0 * INT_2_32_1 + z1 * INT_2_32_0;","    x2 = z1 * INT_2_32_1;","  }","","  if (k &lt; BASE) {","    y0 = k;","    y1 = 0;","  } else {","    y0 = k % BASE;","    y1 = Math.floor(k / BASE);","  }","","  if (l &lt; BASE) {","    y0 += l * INT_2_32_0;","    y1 += l * INT_2_32_1;","    y2 = 0;","  } else {","    z0 = l % BASE;","    z1 = Math.floor(l / BASE);","    y0 += z0 * INT_2_32_0;","    y1 += z0 * INT_2_32_1 + z1 * INT_2_32_0;","    y2 = z1 * INT_2_32_1;","  }","","  if (y0 &gt;= BASE) {","    y1 += Math.floor(y0 / BASE);","    y0 %= BASE;","  }","  if (y1 &gt;= BASE) {","    y2 += Math.floor(y1 / BASE);","    y1 %= BASE;","  }","","  x0 += y0 * INT_2_64_0;","  x1 += y0 * INT_2_64_1 + y1 * INT_2_64_0;","  x2 += y0 * INT_2_64_2 + y1 * INT_2_64_1 + y2 * INT_2_64_0;","  x3 = y1 * INT_2_64_2 + y2 * INT_2_64_1;","  x4 = y2 * INT_2_64_2;","","  if (x0 &gt;= BASE) {","    x1 += Math.floor(x0 / BASE);","    x0 %= BASE;","  }","  if (x1 &gt;= BASE) {","    x2 += Math.floor(x1 / BASE);","    x1 %= BASE;","  }","  if (x2 &gt;= BASE) {","    x3 += Math.floor(x2 / BASE);","    x2 %= BASE;","  }","  if (x3 &gt;= BASE) {","    x4 += Math.floor(x3 / BASE);","    x3 %= BASE;","  }","","  if (x4) {","    dec.m = '' + x4 + zeropad.lpad14(x3 * BASE + x2) + zeropad.lpad14(x1 * BASE +","      x0);","    return dec;","  }","  if (x3) {","    dec.m = '' + (x3 * BASE + x2) + zeropad.lpad14(x1 * BASE + x0);","    return dec;","  }","  if (x2) {","    if (x2 &lt; INT_2_53_2 || (x2 === INT_2_53_2 &amp;&amp; (x1 &lt; INT_2_53_1 || (x1 ===","      INT_2_53_1 &amp;&amp; x0 &lt;= INT_2_53_0)))) {","      dec.m = (x2 * BASE + x1) * BASE + x0;","      return dec;","    }","    dec.m = '' + x2 + zeropad.lpad14(x1 * BASE + x0);","    return dec;","  }","  dec.m = x1 * BASE + x0;","  return dec;","","}","","function readDecFloat(buffer, offset) {","  var value = readDec128(buffer, offset);","  if (value === null) {","    return null;","  }","  var d, l, e;","  d = '' + value.m;","  l = d.length;","  e = value.e + l - 1;","  if (e &lt; 0) {","    e = 'e' + e;","  } else {","    e = 'e+' + e;","  }","  while (d.charCodeAt(l - 1) === 48) {","    l -= 1;","  }","  if (l &gt; 1) {","    d = d.substring(0, 1) + '.' + d.substring(1, l);","  }","  if (value.s &lt; 0) {","    return '-' + d + e;","  }","  return d + e;","}","","function readDecFixed(buffer, offset, frac) {","  var value = readDec128(buffer, offset);","  if (value === null) {","    return null;","  }","  frac = frac || 0;","  var d, e, l, k, i, f;","  d = '' + value.m;","  e = value.e;","  if (e &lt; 0) {","    l = d.length;","    k = l + e;","    if (k &gt; 0) {","      i = d.substring(0, k);","      f = d.substring(k);","    } else if (k &lt; 0) {","      i = '0';","      f = zeropad.ZEROS[-k] + d;","    } else {","      i = '0';","      f = d;","    }","  } else if (e &gt; 0) {","    i = d + zeropad.ZEROS[e];","    f = '';","  } else {","    i = d;","    f = '';","  }","  if (value.s &lt; 0) {","    i = '-' + i;","  }","  if (frac === 0) {","    return i;","  }","  l = f.length;","  if (l &gt; frac) {","    f = f.substring(0, frac);","  } else if (l &lt; frac) {","    f += zeropad.ZEROS[frac - l];","  }","  return i + '.' + f;","}","","","function _writeInt64(buffer, value, offset, unsigned) {","","  var l, x, a, b, c, x0, x1, y0, y1, z0, z1, z2, z3;","","  var negative = false;","","  if (typeof value === 'string') {","","    l = value.length;","","    if (l &gt; 15) {","","      a = +value.substring(l - 7);","      b = +value.substring(l - 14, l - 7);","      c = +value.substring(0, l - 14);","","      if (!unsigned &amp;&amp; c &lt; 0) {","        c *= -1;","        negative = true;","      }","","      z0 = a % INT_2_16;","      z1 = Math.floor(a / INT_2_16);","      z2 = z3 = 0;","","      y0 = b % INT_2_16;","      y1 = Math.floor(b / INT_2_16);","","      z0 += y0 * BIN_10_7_0;","      z1 += y0 * BIN_10_7_1 + y1 * BIN_10_7_0;","      z2 += y1 * BIN_10_7_1;","","      y0 = c % INT_2_16;","      y1 = Math.floor(c / INT_2_16);","","      z0 += y0 * BIN_10_14_0;","      z1 += y0 * BIN_10_14_1 + y1 * BIN_10_14_0;","      z2 += y0 * BIN_10_14_2 + y1 * BIN_10_14_1;","      z3 += y1 * BIN_10_14_2;","","      if (z0 &gt;= INT_2_16) {","        z1 += Math.floor(z0 / INT_2_16);","        z0 %= INT_2_16;","      }","      if (z1 &gt;= INT_2_16) {","        z2 += Math.floor(z1 / INT_2_16);","        z1 %= INT_2_16;","      }","      if (z2 &gt;= INT_2_16) {","        z3 += Math.floor(z2 / INT_2_16);","        z2 %= INT_2_16;","      }","","      x0 = z1 * INT_2_16 + z0;","      x1 = z3 * INT_2_16 + z2;","","    } else {","      x = +value;","","      if (!unsigned &amp;&amp; x &lt; 0) {","        x *= -1;","        negative = true;","      }","","      x0 = x % INT_2_32;","      x1 = Math.floor(x / INT_2_32);","    }","  } else {","    x = value;","","    if (!unsigned &amp;&amp; x &lt; 0) {","      x *= -1;","      negative = true;","    }","","    x0 = x % INT_2_32;","    x1 = Math.floor(x / INT_2_32);","  }","","  if (negative) {","    x1 = 0xffffffff - x1;","    if (x0 === 0) {","      x1 += 1;","    } else {","      x0 = 0xffffffff - x0 + 1;","    }","  }","","  buffer[offset + 3] = (x0 &gt;&gt;&gt; 24) &amp; 0xff;","  buffer[offset + 2] = (x0 &gt;&gt;&gt; 16) &amp; 0xff;","  buffer[offset + 1] = (x0 &gt;&gt;&gt; 8) &amp; 0xff;","  buffer[offset] = x0 &amp; 0xff;","  offset += 4;","","  buffer[offset + 3] = (x1 &gt;&gt;&gt; 24) &amp; 0xff;","  buffer[offset + 2] = (x1 &gt;&gt;&gt; 16) &amp; 0xff;","  buffer[offset + 1] = (x1 &gt;&gt;&gt; 8) &amp; 0xff;","  buffer[offset] = x1 &amp; 0xff;","}","","function readUInt128(buffer, offset) {","","  var i, j, k, l, z0, z1, y0, y1, y2, x0, x1, x2, x3, x4;","","  offset = offset || 0;","","  i = buffer[offset + 2] &lt;&lt; 16;","  i |= buffer[offset + 1] &lt;&lt; 8;","  i |= buffer[offset];","  i += buffer[offset + 3] &lt;&lt; 24 &gt;&gt;&gt; 0;","  offset += 4;","","  j = buffer[offset + 2] &lt;&lt; 16;","  j |= buffer[offset + 1] &lt;&lt; 8;","  j |= buffer[offset];","  j += buffer[offset + 3] &lt;&lt; 24 &gt;&gt;&gt; 0;","  offset += 4;","","  k = buffer[offset + 2] &lt;&lt; 16;","  k |= buffer[offset + 1] &lt;&lt; 8;","  k |= buffer[offset];","  k += buffer[offset + 3] &lt;&lt; 24 &gt;&gt;&gt; 0;","  offset += 4;","","  l = (buffer[offset + 2]) &lt;&lt; 16;","  l |= buffer[offset + 1] &lt;&lt; 8;","  l |= buffer[offset];","  l += buffer[offset + 3] &lt;&lt; 24 &gt;&gt;&gt; 0;","","  if (k === 0 &amp;&amp; l === 0) {","    if (j === 0) {","      return i;","    }","    if (j &lt; INT_2_21 || (j === INT_2_21 &amp;&amp; i === 0)) {","      return j * INT_2_32 + i;","    }","  }","","  if (i &lt; BASE) {","    x0 = i;","    x1 = 0;","  } else {","    x0 = i % BASE;","    x1 = Math.floor(i / BASE);","  }","","  if (j &lt; BASE) {","    x0 += j * INT_2_32_0;","    x1 += j * INT_2_32_1;","    x2 = 0;","  } else {","    z0 = j % BASE;","    z1 = Math.floor(j / BASE);","    x0 += z0 * INT_2_32_0;","    x1 += z0 * INT_2_32_1 + z1 * INT_2_32_0;","    x2 = z1 * INT_2_32_1;","  }","","  if (k &lt; BASE) {","    y0 = k;","    y1 = 0;","  } else {","    y0 = k % BASE;","    y1 = Math.floor(k / BASE);","  }","","  if (l &lt; BASE) {","    y0 += l * INT_2_32_0;","    y1 += l * INT_2_32_1;","    y2 = 0;","  } else {","    z0 = l % BASE;","    z1 = Math.floor(l / BASE);","    y0 += z0 * INT_2_32_0;","    y1 += z0 * INT_2_32_1 + z1 * INT_2_32_0;","    y2 = z1 * INT_2_32_1;","  }","","  if (y0 &gt;= BASE) {","    y1 += Math.floor(y0 / BASE);","    y0 %= BASE;","  }","  if (y1 &gt;= BASE) {","    y2 += Math.floor(y1 / BASE);","    y1 %= BASE;","  }","","  x0 += y0 * INT_2_64_0;","  x1 += y0 * INT_2_64_1 + y1 * INT_2_64_0;","  x2 += y0 * INT_2_64_2 + y1 * INT_2_64_1 + y2 * INT_2_64_0;","  x3 = y1 * INT_2_64_2 + y2 * INT_2_64_1;","  x4 = y2 * INT_2_64_2;","","  if (x0 &gt;= BASE) {","    x1 += Math.floor(x0 / BASE);","    x0 %= BASE;","  }","  if (x1 &gt;= BASE) {","    x2 += Math.floor(x1 / BASE);","    x1 %= BASE;","  }","  if (x2 &gt;= BASE) {","    x3 += Math.floor(x2 / BASE);","    x2 %= BASE;","  }","  if (x3 &gt;= BASE) {","    x4 += Math.floor(x3 / BASE);","    x3 %= BASE;","  }","","  if (x4) {","    return '' + x4 + zeropad.lpad14(x3 * BASE + x2) + zeropad.lpad14(x1 * BASE +","      x0);","  }","  if (x3) {","    return '' + (x3 * BASE + x2) + zeropad.lpad14(x1 * BASE + x0);","  }","  if (x2) {","    if (x2 &lt; INT_2_53_2 || (x2 === INT_2_53_2 &amp;&amp;","      (x1 &lt; INT_2_53_1 || (x1 === INT_2_53_1 &amp;&amp; x0 &lt;= INT_2_53_0)))) {","      return (x2 * BASE + x1) * BASE + x0;","    }","    return '' + x2 + zeropad.lpad14(x1 * BASE + x0);","","  }","  return x1 * BASE + x0;","}","","","function writeUInt128(buffer, value, offset) {","","  var l, a, b, c, d, e, f,","    x0, x1, x2, x3, y0, y1, z0, z1, z2, z3, z4, z5, z6, z7;","","  l = value.length;","","  a = +value.substring(l - 7);","  if (l &gt; 7) {","    b = +value.substring(l - 14, l - 7);","  } else {","    b = 0;","  }","  if (l &gt; 14) {","    c = +value.substring(l - 21, l - 14);","  } else {","    c = 0;","  }","  if (l &gt; 21) {","    d = +value.substring(l - 28, l - 21);","  } else {","    d = 0;","  }","  if (l &gt; 28) {","    e = +value.substring(l - 35, l - 28);","  } else {","    e = 0;","  }","  if (l &gt; 35) {","    f = +value.substring(0, l - 35);","  } else {","    f = 0;","  }","","  z0 = z1 = z2 = z3 = z4 = z5 = z6 = z7 = 0;","","  // set a * 10^0","  if (a) {","    z0 = a % INT_2_16;","    z1 = Math.floor(a / INT_2_16);","  }","","  // add b * 10^7","  if (b) {","    y0 = b % INT_2_16;","    y1 = Math.floor(b / INT_2_16);","    z0 += y0 * BIN_10_7_0;","    z1 += y0 * BIN_10_7_1 + y1 * BIN_10_7_0;","    z2 += y1 * BIN_10_7_1;","  }","","  // add c * 10^14","  if (c) {","    y0 = c % INT_2_16;","    y1 = Math.floor(c / INT_2_16);","    z0 += y0 * BIN_10_14_0;","    z1 += y0 * BIN_10_14_1 + y1 * BIN_10_14_0;","    z2 += y0 * BIN_10_14_2 + y1 * BIN_10_14_1;","    z3 += y1 * BIN_10_14_2;","  }","","  // add d * 10^21","  if (d) {","    y0 = d % INT_2_16;","    y1 = Math.floor(d / INT_2_16);","    z0 += y0 * BIN_10_21_0;","    z1 += y0 * BIN_10_21_1 + y1 * BIN_10_21_0;","    z2 += y0 * BIN_10_21_2 + y1 * BIN_10_21_1;","    z3 += y0 * BIN_10_21_3 + y1 * BIN_10_21_2;","    z4 += y0 * BIN_10_21_4 + y1 * BIN_10_21_3;","    z5 += y1 * BIN_10_21_4;","  }","","  // add e * 10^28","  if (e) {","    y0 = e % INT_2_16;","    y1 = Math.floor(e / INT_2_16);","    z0 += y0 * BIN_10_28_0;","    z1 += y0 * BIN_10_28_1 + y1 * BIN_10_28_0;","    z2 += y0 * BIN_10_28_2 + y1 * BIN_10_28_1;","    z3 += y0 * BIN_10_28_3 + y1 * BIN_10_28_2;","    z4 += y0 * BIN_10_28_4 + y1 * BIN_10_28_3;","    z5 += y0 * BIN_10_28_5 + y1 * BIN_10_28_4;","    z6 += y1 * BIN_10_28_5;","  }","","  // add f * 10^35","  if (f) {","    y0 = f % INT_2_16;","    z0 += y0 * BIN_10_35_0;","    z1 += y0 * BIN_10_35_1;","    z2 += y0 * BIN_10_35_2;","    z3 += y0 * BIN_10_35_3;","    z4 += y0 * BIN_10_35_4;","    z5 += y0 * BIN_10_35_5;","    z6 += y0 * BIN_10_35_6;","    z7 += y0 * BIN_10_35_7;","  }","","  if (z0 &gt;= INT_2_16) {","    z1 += Math.floor(z0 / INT_2_16);","    z0 %= INT_2_16;","  }","  if (z1 &gt;= INT_2_16) {","    z2 += Math.floor(z1 / INT_2_16);","    z1 %= INT_2_16;","  }","  if (z2 &gt;= INT_2_16) {","    z3 += Math.floor(z2 / INT_2_16);","    z2 %= INT_2_16;","  }","  if (z3 &gt;= INT_2_16) {","    z4 += Math.floor(z3 / INT_2_16);","    z3 %= INT_2_16;","  }","  if (z4 &gt;= INT_2_16) {","    z5 += Math.floor(z4 / INT_2_16);","    z4 %= INT_2_16;","  }","  if (z5 &gt;= INT_2_16) {","    z6 += Math.floor(z5 / INT_2_16);","    z5 %= INT_2_16;","  }","  if (z6 &gt;= INT_2_16) {","    z7 += Math.floor(z6 / INT_2_16);","    z6 %= INT_2_16;","  }","","  x0 = z1 * INT_2_16 + z0;","  x1 = z3 * INT_2_16 + z2;","  x2 = z5 * INT_2_16 + z4;","  x3 = z7 * INT_2_16 + z6;","","  buffer[offset + 3] = (x0 &gt;&gt;&gt; 24) &amp; 0xff;","  buffer[offset + 2] = (x0 &gt;&gt;&gt; 16) &amp; 0xff;","  buffer[offset + 1] = (x0 &gt;&gt;&gt; 8) &amp; 0xff;","  buffer[offset] = x0 &amp; 0xff;","  offset += 4;","","  buffer[offset + 3] = (x1 &gt;&gt;&gt; 24) &amp; 0xff;","  buffer[offset + 2] = (x1 &gt;&gt;&gt; 16) &amp; 0xff;","  buffer[offset + 1] = (x1 &gt;&gt;&gt; 8) &amp; 0xff;","  buffer[offset] = x1 &amp; 0xff;","  offset += 4;","","  buffer[offset + 3] = (x2 &gt;&gt;&gt; 24) &amp; 0xff;","  buffer[offset + 2] = (x2 &gt;&gt;&gt; 16) &amp; 0xff;","  buffer[offset + 1] = (x2 &gt;&gt;&gt; 8) &amp; 0xff;","  buffer[offset] = x2 &amp; 0xff;","  offset += 4;","","  buffer[offset + 3] = (x3 &gt;&gt;&gt; 24) &amp; 0xff;","  buffer[offset + 2] = (x3 &gt;&gt;&gt; 16) &amp; 0xff;","  buffer[offset + 1] = (x3 &gt;&gt;&gt; 8) &amp; 0xff;","  buffer[offset] = x3 &amp; 0xff;","}","","function readInt64(buffer, offset) {","  return _readInt64(buffer, offset || 0, false);","}","","function readUInt64(buffer, offset) {","  return _readInt64(buffer, offset || 0, true);","}","","function writeInt64(buffer, value, offset) {","  _writeInt64(buffer, value, offset || 0, false);","}","","function writeUInt64(buffer, value, offset) {","  _writeInt64(buffer, value, offset || 0, true);","}","","exports.readInt64LE = readInt64;","exports.readUInt64LE = readUInt64;","exports.writeInt64LE = writeInt64;","exports.writeUInt64LE = writeUInt64;","exports.readUInt128LE = readUInt128;","exports.writeUInt128LE = writeUInt128;","exports.readDec128 = readDec128;","exports.writeDec128 = writeDec128;","exports.readDecFloat = readDecFloat;","exports.readDecFixed = readDecFixed;"];
