// hdb/lib-cov/util/index.js
/* automatically generated by JSCoverage - do not edit */
if (typeof _$jscoverage === 'undefined') _$jscoverage = {};
if (! _$jscoverage['util/index.js']) {
  _$jscoverage['util/index.js'] = [];
  _$jscoverage['util/index.js'][14] = 0;
  _$jscoverage['util/index.js'][18] = 0;
  _$jscoverage['util/index.js'][19] = 0;
  _$jscoverage['util/index.js'][20] = 0;
  _$jscoverage['util/index.js'][21] = 0;
  _$jscoverage['util/index.js'][22] = 0;
  _$jscoverage['util/index.js'][24] = 0;
  _$jscoverage['util/index.js'][26] = 0;
  _$jscoverage['util/index.js'][28] = 0;
  _$jscoverage['util/index.js'][31] = 0;
  _$jscoverage['util/index.js'][33] = 0;
  _$jscoverage['util/index.js'][35] = 0;
  _$jscoverage['util/index.js'][42] = 0;
  _$jscoverage['util/index.js'][43] = 0;
  _$jscoverage['util/index.js'][46] = 0;
  _$jscoverage['util/index.js'][61] = 0;
  _$jscoverage['util/index.js'][62] = 0;
  _$jscoverage['util/index.js'][63] = 0;
  _$jscoverage['util/index.js'][64] = 0;
  _$jscoverage['util/index.js'][66] = 0;
  _$jscoverage['util/index.js'][67] = 0;
  _$jscoverage['util/index.js'][68] = 0;
  _$jscoverage['util/index.js'][70] = 0;
  _$jscoverage['util/index.js'][71] = 0;
  _$jscoverage['util/index.js'][74] = 0;
  _$jscoverage['util/index.js'][75] = 0;
  _$jscoverage['util/index.js'][79] = 0;
  _$jscoverage['util/index.js'][81] = 0;
  _$jscoverage['util/index.js'][82] = 0;
  _$jscoverage['util/index.js'][83] = 0;
  _$jscoverage['util/index.js'][84] = 0;
  _$jscoverage['util/index.js'][85] = 0;
  _$jscoverage['util/index.js'][86] = 0;
  _$jscoverage['util/index.js'][87] = 0;
  _$jscoverage['util/index.js'][88] = 0;
  _$jscoverage['util/index.js'][89] = 0;
  _$jscoverage['util/index.js'][96] = 0;
  _$jscoverage['util/index.js'][99] = 0;
  _$jscoverage['util/index.js'][102] = 0;
  _$jscoverage['util/index.js'][103] = 0;
  _$jscoverage['util/index.js'][105] = 0;
  _$jscoverage['util/index.js'][106] = 0;
  _$jscoverage['util/index.js'][107] = 0;
  _$jscoverage['util/index.js'][111] = 0;
  _$jscoverage['util/index.js'][112] = 0;
  _$jscoverage['util/index.js'][114] = 0;
  _$jscoverage['util/index.js'][116] = 0;
  _$jscoverage['util/index.js'][117] = 0;
  _$jscoverage['util/index.js'][119] = 0;
  _$jscoverage['util/index.js'][121] = 0;
  _$jscoverage['util/index.js'][122] = 0;
  _$jscoverage['util/index.js'][124] = 0;
  _$jscoverage['util/index.js'][126] = 0;
  _$jscoverage['util/index.js'][127] = 0;
  _$jscoverage['util/index.js'][129] = 0;
  _$jscoverage['util/index.js'][131] = 0;
  _$jscoverage['util/index.js'][132] = 0;
  _$jscoverage['util/index.js'][134] = 0;
  _$jscoverage['util/index.js'][136] = 0;
  _$jscoverage['util/index.js'][137] = 0;
  _$jscoverage['util/index.js'][139] = 0;
  _$jscoverage['util/index.js'][141] = 0;
  _$jscoverage['util/index.js'][142] = 0;
  _$jscoverage['util/index.js'][144] = 0;
  _$jscoverage['util/index.js'][146] = 0;
  _$jscoverage['util/index.js'][147] = 0;
  _$jscoverage['util/index.js'][148] = 0;
  _$jscoverage['util/index.js'][150] = 0;
  _$jscoverage['util/index.js'][152] = 0;
  _$jscoverage['util/index.js'][154] = 0;
  _$jscoverage['util/index.js'][155] = 0;
  _$jscoverage['util/index.js'][157] = 0;
  _$jscoverage['util/index.js'][159] = 0;
  _$jscoverage['util/index.js'][160] = 0;
  _$jscoverage['util/index.js'][161] = 0;
  _$jscoverage['util/index.js'][164] = 0;
  _$jscoverage['util/index.js'][166] = 0;
  _$jscoverage['util/index.js'][167] = 0;
  _$jscoverage['util/index.js'][168] = 0;
  _$jscoverage['util/index.js'][170] = 0;
  _$jscoverage['util/index.js'][171] = 0;
  _$jscoverage['util/index.js'][172] = 0;
  _$jscoverage['util/index.js'][173] = 0;
  _$jscoverage['util/index.js'][174] = 0;
  _$jscoverage['util/index.js'][175] = 0;
  _$jscoverage['util/index.js'][176] = 0;
  _$jscoverage['util/index.js'][178] = 0;
  _$jscoverage['util/index.js'][182] = 0;
  _$jscoverage['util/index.js'][183] = 0;
  _$jscoverage['util/index.js'][185] = 0;
  _$jscoverage['util/index.js'][187] = 0;
  _$jscoverage['util/index.js'][188] = 0;
  _$jscoverage['util/index.js'][189] = 0;
  _$jscoverage['util/index.js'][191] = 0;
  _$jscoverage['util/index.js'][193] = 0;
  _$jscoverage['util/index.js'][194] = 0;
  _$jscoverage['util/index.js'][196] = 0;
  _$jscoverage['util/index.js'][197] = 0;
  _$jscoverage['util/index.js'][199] = 0;
  _$jscoverage['util/index.js'][201] = 0;
  _$jscoverage['util/index.js'][203] = 0;
  _$jscoverage['util/index.js'][204] = 0;
  _$jscoverage['util/index.js'][206] = 0;
  _$jscoverage['util/index.js'][207] = 0;
  _$jscoverage['util/index.js'][209] = 0;
  _$jscoverage['util/index.js'][211] = 0;
  _$jscoverage['util/index.js'][212] = 0;
  _$jscoverage['util/index.js'][213] = 0;
  _$jscoverage['util/index.js'][214] = 0;
  _$jscoverage['util/index.js'][217] = 0;
  _$jscoverage['util/index.js'][218] = 0;
  _$jscoverage['util/index.js'][219] = 0;
  _$jscoverage['util/index.js'][222] = 0;
  _$jscoverage['util/index.js'][223] = 0;
  _$jscoverage['util/index.js'][225] = 0;
  _$jscoverage['util/index.js'][227] = 0;
  _$jscoverage['util/index.js'][228] = 0;
  _$jscoverage['util/index.js'][229] = 0;
  _$jscoverage['util/index.js'][232] = 0;
  _$jscoverage['util/index.js'][233] = 0;
  _$jscoverage['util/index.js'][236] = 0;
  _$jscoverage['util/index.js'][238] = 0;
  _$jscoverage['util/index.js'][239] = 0;
  _$jscoverage['util/index.js'][241] = 0;
  _$jscoverage['util/index.js'][243] = 0;
  _$jscoverage['util/index.js'][244] = 0;
  _$jscoverage['util/index.js'][245] = 0;
  _$jscoverage['util/index.js'][247] = 0;
  _$jscoverage['util/index.js'][249] = 0;
  _$jscoverage['util/index.js'][250] = 0;
  _$jscoverage['util/index.js'][251] = 0;
  _$jscoverage['util/index.js'][252] = 0;
  _$jscoverage['util/index.js'][253] = 0;
  _$jscoverage['util/index.js'][254] = 0;
}
_$jscoverage['util/index.js'][14]++;
"use strict";
_$jscoverage['util/index.js'][18]++;
var util = require("util");
_$jscoverage['util/index.js'][19]++;
var stream = require("stream");
_$jscoverage['util/index.js'][20]++;
var path = require("path");
_$jscoverage['util/index.js'][21]++;
var os = require("os");
_$jscoverage['util/index.js'][22]++;
var fs = require("fs");
_$jscoverage['util/index.js'][24]++;
if (typeof setImmediate === "undefined") {
  _$jscoverage['util/index.js'][26]++;
  exports.setImmediate = process.nextTick;
}
else {
  _$jscoverage['util/index.js'][28]++;
  exports.setImmediate = setImmediate;
}
_$jscoverage['util/index.js'][31]++;
if (! stream.Readable) {
  _$jscoverage['util/index.js'][33]++;
  stream = require("readable-stream");
}
_$jscoverage['util/index.js'][35]++;
exports.stream = {Readable: stream.Readable, Writable: stream.Writable, Duplex: stream.Duplex, Transform: stream.Transform};
_$jscoverage['util/index.js'][42]++;
function exportNativeUtil(fn) {
  _$jscoverage['util/index.js'][43]++;
  exports[fn] = util[fn];
}
_$jscoverage['util/index.js'][46]++;
["format", "debug", "error", "puts", "print", "log", "inspect", "isArray", "isRegExp", "isDate", "isError", "inherits"].forEach(exportNativeUtil);
_$jscoverage['util/index.js'][61]++;
exports.bignum = require("./bignum");
_$jscoverage['util/index.js'][62]++;
exports.calendar = require("./calendar");
_$jscoverage['util/index.js'][63]++;
exports.Queue = require("./Queue");
_$jscoverage['util/index.js'][64]++;
extend(exports, require("./zeropad"));
_$jscoverage['util/index.js'][66]++;
var debuglog;
_$jscoverage['util/index.js'][67]++;
if (util.debuglog) {
  _$jscoverage['util/index.js'][68]++;
  debuglog = util.debuglog;
}
else {
  _$jscoverage['util/index.js'][70]++;
  try {
    _$jscoverage['util/index.js'][71]++;
    debuglog = require("debuglog");
  }
  catch (err) {
    _$jscoverage['util/index.js'][74]++;
    debuglog = (function dbglg(name) {
  _$jscoverage['util/index.js'][75]++;
  return (function () {
});
});
  }
}
_$jscoverage['util/index.js'][79]++;
exports.debuglog = debuglog;
_$jscoverage['util/index.js'][81]++;
var traceFunction;
_$jscoverage['util/index.js'][82]++;
exports.tracelog = (function tracelog() {
  _$jscoverage['util/index.js'][83]++;
  if (typeof traceFunction !== "function") {
    _$jscoverage['util/index.js'][84]++;
    if (process.env.HDB_TRACE) {
      _$jscoverage['util/index.js'][85]++;
      var timestamp = Math.floor(Date.now() / 1000);
      _$jscoverage['util/index.js'][86]++;
      var filename = path.join(os.tmpDir(), "hdb.trace." + timestamp + ".log");
      _$jscoverage['util/index.js'][87]++;
      console.log("Trace to file", filename);
      _$jscoverage['util/index.js'][88]++;
      traceFunction = (function trace(kind, segment) {
  _$jscoverage['util/index.js'][89]++;
  fs.appendFileSync(filename, kind + "\n" + util.inspect(segment, {depth: 9}));
});
    }
    else {
      _$jscoverage['util/index.js'][96]++;
      traceFunction = (function trace() {
});
    }
  }
  _$jscoverage['util/index.js'][99]++;
  return traceFunction;
});
_$jscoverage['util/index.js'][102]++;
function extend(obj) {
  _$jscoverage['util/index.js'][103]++;
  function extendOnce(source) {
    _$jscoverage['util/index.js'][105]++;
    if (source) {
      _$jscoverage['util/index.js'][106]++;
      for (var prop in source) {
        _$jscoverage['util/index.js'][107]++;
        obj[prop] = source[prop];
}
    }
}
  _$jscoverage['util/index.js'][111]++;
  Array.prototype.slice.call(arguments, 1).forEach(extendOnce);
  _$jscoverage['util/index.js'][112]++;
  return obj;
}
_$jscoverage['util/index.js'][114]++;
exports.extend = extend;
_$jscoverage['util/index.js'][116]++;
function isBoolean(arg) {
  _$jscoverage['util/index.js'][117]++;
  return typeof arg === "boolean";
}
_$jscoverage['util/index.js'][119]++;
exports.isBoolean = isBoolean;
_$jscoverage['util/index.js'][121]++;
function isNumber(arg) {
  _$jscoverage['util/index.js'][122]++;
  return typeof arg === "number";
}
_$jscoverage['util/index.js'][124]++;
exports.isNumber = isNumber;
_$jscoverage['util/index.js'][126]++;
function isString(arg) {
  _$jscoverage['util/index.js'][127]++;
  return typeof arg === "string";
}
_$jscoverage['util/index.js'][129]++;
exports.isString = isString;
_$jscoverage['util/index.js'][131]++;
function isUndefined(arg) {
  _$jscoverage['util/index.js'][132]++;
  return arg === void 0;
}
_$jscoverage['util/index.js'][134]++;
exports.isUndefined = isUndefined;
_$jscoverage['util/index.js'][136]++;
function isObject(arg) {
  _$jscoverage['util/index.js'][137]++;
  return typeof arg === "object" && arg !== null;
}
_$jscoverage['util/index.js'][139]++;
exports.isObject = isObject;
_$jscoverage['util/index.js'][141]++;
function isFunction(arg) {
  _$jscoverage['util/index.js'][142]++;
  return typeof arg === "function";
}
_$jscoverage['util/index.js'][144]++;
exports.isFunction = isFunction;
_$jscoverage['util/index.js'][146]++;
function alignLength(length, alignment) {
  _$jscoverage['util/index.js'][147]++;
  if (length % alignment === 0) {
    _$jscoverage['util/index.js'][148]++;
    return length;
  }
  _$jscoverage['util/index.js'][150]++;
  return length + alignment - length % alignment;
}
_$jscoverage['util/index.js'][152]++;
exports.alignLength = alignLength;
_$jscoverage['util/index.js'][154]++;
function cc2_(str) {
  _$jscoverage['util/index.js'][155]++;
  return str.replace(/([a-z])([A-Z])/g, "$1_$2").toUpperCase();
}
_$jscoverage['util/index.js'][157]++;
exports.cc2_ = cc2_;
_$jscoverage['util/index.js'][159]++;
function _2cc(str) {
  _$jscoverage['util/index.js'][160]++;
  return str.toLowerCase().replace(/_([a-z])/g, (function toUpperCase(match, str) {
  _$jscoverage['util/index.js'][161]++;
  return str.toUpperCase();
}));
}
_$jscoverage['util/index.js'][164]++;
exports._2cc = _2cc;
_$jscoverage['util/index.js'][166]++;
function readData(ds, cb) {
  _$jscoverage['util/index.js'][167]++;
  var length = 0;
  _$jscoverage['util/index.js'][168]++;
  var chunks = [];
  _$jscoverage['util/index.js'][170]++;
  function done(err) {
    _$jscoverage['util/index.js'][171]++;
    ds.removeListener("error", onerror);
    _$jscoverage['util/index.js'][172]++;
    ds.removeListener("data", ondata);
    _$jscoverage['util/index.js'][173]++;
    ds.removeListener("end", onend);
    _$jscoverage['util/index.js'][174]++;
    if (isFunction(cb)) {
      _$jscoverage['util/index.js'][175]++;
      if (err) {
        _$jscoverage['util/index.js'][176]++;
        return cb(err);
      }
      _$jscoverage['util/index.js'][178]++;
      cb(null, Buffer.concat(chunks, length));
    }
}
  _$jscoverage['util/index.js'][182]++;
  function onerror(err) {
    _$jscoverage['util/index.js'][183]++;
    done(err);
}
  _$jscoverage['util/index.js'][185]++;
  ds.on("error", onerror);
  _$jscoverage['util/index.js'][187]++;
  function ondata(chunk) {
    _$jscoverage['util/index.js'][188]++;
    chunks.push(chunk);
    _$jscoverage['util/index.js'][189]++;
    length += chunk.length;
}
  _$jscoverage['util/index.js'][191]++;
  ds.on("data", ondata);
  _$jscoverage['util/index.js'][193]++;
  function onend() {
    _$jscoverage['util/index.js'][194]++;
    done(null);
}
  _$jscoverage['util/index.js'][196]++;
  ds.on("end", onend);
  _$jscoverage['util/index.js'][197]++;
  ds.resume();
}
_$jscoverage['util/index.js'][199]++;
exports.readData = readData;
_$jscoverage['util/index.js'][201]++;
function proxyEvents(source, target, events) {
  _$jscoverage['util/index.js'][203]++;
  function proxyEvent(ev) {
    _$jscoverage['util/index.js'][204]++;
    source.on(ev, target.emit.bind(target, ev));
}
  _$jscoverage['util/index.js'][206]++;
  events.forEach(proxyEvent);
  _$jscoverage['util/index.js'][207]++;
  return target;
}
_$jscoverage['util/index.js'][209]++;
exports.proxyEvents = proxyEvents;
_$jscoverage['util/index.js'][211]++;
function createReadStream(ds, events, options) {
  _$jscoverage['util/index.js'][212]++;
  if (! util.isArray(events)) {
    _$jscoverage['util/index.js'][213]++;
    options = events;
    _$jscoverage['util/index.js'][214]++;
    events = ["error", "close"];
  }
  _$jscoverage['util/index.js'][217]++;
  var readable = new stream.Readable(options);
  _$jscoverage['util/index.js'][218]++;
  readable._read = (function _read() {
  _$jscoverage['util/index.js'][219]++;
  ds.resume();
});
  _$jscoverage['util/index.js'][222]++;
  function onend() {
    _$jscoverage['util/index.js'][223]++;
    readable.push(null);
}
  _$jscoverage['util/index.js'][225]++;
  ds.once("end", onend);
  _$jscoverage['util/index.js'][227]++;
  function ondata(chunk) {
    _$jscoverage['util/index.js'][228]++;
    if (! chunk || ! chunk.length) {
      _$jscoverage['util/index.js'][229]++;
      return;
    }
    _$jscoverage['util/index.js'][232]++;
    if (! readable.push(chunk)) {
      _$jscoverage['util/index.js'][233]++;
      ds.pause();
    }
}
  _$jscoverage['util/index.js'][236]++;
  ds.on("data", ondata);
  _$jscoverage['util/index.js'][238]++;
  proxyEvents(ds, readable, events);
  _$jscoverage['util/index.js'][239]++;
  return readable;
}
_$jscoverage['util/index.js'][241]++;
exports.createReadStream = createReadStream;
_$jscoverage['util/index.js'][243]++;
function pipe(source, target, events) {
  _$jscoverage['util/index.js'][244]++;
  proxyEvents(source, target, events || ["error"]);
  _$jscoverage['util/index.js'][245]++;
  return source.pipe(target);
}
_$jscoverage['util/index.js'][247]++;
exports.pipe = pipe;
_$jscoverage['util/index.js'][249]++;
function createReadableBuffer(buffer, options) {
  _$jscoverage['util/index.js'][250]++;
  options = options || {};
  _$jscoverage['util/index.js'][251]++;
  var readable = new stream.Readable(options);
  _$jscoverage['util/index.js'][252]++;
  readable.push(buffer);
  _$jscoverage['util/index.js'][253]++;
  readable.push(null);
  _$jscoverage['util/index.js'][254]++;
  return readable;
}
_$jscoverage['util/index.js'].source = ["// Copyright 2013 SAP AG.","//","// Licensed under the Apache License, Version 2.0 (the \"License\");","// you may not use this file except in compliance with the License.","// You may obtain a copy of the License at","//","// http: //www.apache.org/licenses/LICENSE-2.0","//","// Unless required by applicable law or agreed to in writing,","// software distributed under the License is distributed on an","// \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,","// either express or implied. See the License for the specific","// language governing permissions and limitations under the License.","'use strict';","","/* jshint camelcase:false */","","var util = require('util');","var stream = require('stream');","var path = require('path');","var os = require('os');","var fs = require('fs');","","if (typeof setImmediate === 'undefined') {","  // inoffical 0.8.x support","  exports.setImmediate = process.nextTick;","} else {","  exports.setImmediate = setImmediate;","}","","if (!stream.Readable) {","  // inoffical 0.8.x support","  stream = require('readable-stream');","}","exports.stream = {","  Readable: stream.Readable,","  Writable: stream.Writable,","  Duplex: stream.Duplex,","  Transform: stream.Transform","};","","function exportNativeUtil(fn) {","  exports[fn] = util[fn];","}","","[","  'format',","  'debug',","  'error',","  'puts',","  'print',","  'log',","  'inspect',","  'isArray',","  'isRegExp',","  'isDate',","  'isError',","  'inherits'","].forEach(exportNativeUtil);","","exports.bignum = require('./bignum');","exports.calendar = require('./calendar');","exports.Queue = require('./Queue');","extend(exports, require('./zeropad'));","","var debuglog;","if (util.debuglog) {","  debuglog = util.debuglog;","} else {","  try {","    debuglog = require('debuglog');","  } catch (err) {","    /* jshint unused:false */","    debuglog = function dbglg(name) {","      return function () {};","    };","  }","}","exports.debuglog = debuglog;","","var traceFunction;","exports.tracelog = function tracelog() {","  if (typeof traceFunction !== 'function') {","    if (process.env.HDB_TRACE) {","      var timestamp = Math.floor(Date.now() / 1000);","      var filename = path.join(os.tmpDir(), 'hdb.trace.' + timestamp + '.log');","      console.log('Trace to file', filename);","      traceFunction = function trace(kind, segment) {","        fs.appendFileSync(filename,","          kind + '\\n' +","          util.inspect(segment, {","            depth: 9","          }));","      };","    } else {","      traceFunction = function trace() {};","    }","  }","  return traceFunction;","};","","function extend(obj) {","  function extendOnce(source) {","    /* jshint forin:false */","    if (source) {","      for (var prop in source) {","        obj[prop] = source[prop];","      }","    }","  }","  Array.prototype.slice.call(arguments, 1).forEach(extendOnce);","  return obj;","}","exports.extend = extend;","","function isBoolean(arg) {","  return typeof arg === 'boolean';","}","exports.isBoolean = isBoolean;","","function isNumber(arg) {","  return typeof arg === 'number';","}","exports.isNumber = isNumber;","","function isString(arg) {","  return typeof arg === 'string';","}","exports.isString = isString;","","function isUndefined(arg) {","  return arg === void 0;","}","exports.isUndefined = isUndefined;","","function isObject(arg) {","  return typeof arg === 'object' &amp;&amp; arg !== null;","}","exports.isObject = isObject;","","function isFunction(arg) {","  return typeof arg === 'function';","}","exports.isFunction = isFunction;","","function alignLength(length, alignment) {","  if (length % alignment === 0) {","    return length;","  }","  return length + alignment - length % alignment;","}","exports.alignLength = alignLength;","","function cc2_(str) {","  return str.replace(/([a-z])([A-Z])/g, '$1_$2').toUpperCase();","}","exports.cc2_ = cc2_;","","function _2cc(str) {","  return str.toLowerCase().replace(/_([a-z])/g, function toUpperCase(match, str) {","    return str.toUpperCase();","  });","}","exports._2cc = _2cc;","","function readData(ds, cb) {","  var length = 0;","  var chunks = [];","","  function done(err) {","    ds.removeListener('error', onerror);","    ds.removeListener('data', ondata);","    ds.removeListener('end', onend);","    if (isFunction(cb)) {","      if (err) {","        return cb(err);","      }","      cb(null, Buffer.concat(chunks, length));","    }","  }","","  function onerror(err) {","    done(err);","  }","  ds.on('error', onerror);","","  function ondata(chunk) {","    chunks.push(chunk);","    length += chunk.length;","  }","  ds.on('data', ondata);","","  function onend() {","    done(null);","  }","  ds.on('end', onend);","  ds.resume();","}","exports.readData = readData;","","function proxyEvents(source, target, events) {","","  function proxyEvent(ev) {","    source.on(ev, target.emit.bind(target, ev));","  }","  events.forEach(proxyEvent);","  return target;","}","exports.proxyEvents = proxyEvents;","","function createReadStream(ds, events, options) {","  if (!util.isArray(events)) {","    options = events;","    events = ['error', 'close'];","  }","","  var readable = new stream.Readable(options);","  readable._read = function _read() {","    ds.resume();","  };","","  function onend() {","    readable.push(null);","  }","  ds.once('end', onend);","","  function ondata(chunk) {","    if (!chunk || !chunk.length) {","      return;","    }","","    if (!readable.push(chunk)) {","      ds.pause();","    }","  }","  ds.on('data', ondata);","","  proxyEvents(ds, readable, events);","  return readable;","}","exports.createReadStream = createReadStream;","","function pipe(source, target, events) {","  proxyEvents(source, target, events || ['error']);","  return source.pipe(target);","}","exports.pipe = pipe;","","function createReadableBuffer(buffer, options) {","  options = options || {};","  var readable = new stream.Readable(options);","  readable.push(buffer);","  readable.push(null);","  return readable;","}"];
