// hdb/lib-cov/util/zeropad.js
/* automatically generated by JSCoverage - do not edit */
if (typeof _$jscoverage === 'undefined') _$jscoverage = {};
if (! _$jscoverage['util/zeropad.js']) {
  _$jscoverage['util/zeropad.js'] = [];
  _$jscoverage['util/zeropad.js'][14] = 0;
  _$jscoverage['util/zeropad.js'][17] = 0;
  _$jscoverage['util/zeropad.js'][18] = 0;
  _$jscoverage['util/zeropad.js'][19] = 0;
  _$jscoverage['util/zeropad.js'][20] = 0;
  _$jscoverage['util/zeropad.js'][21] = 0;
  _$jscoverage['util/zeropad.js'][22] = 0;
  _$jscoverage['util/zeropad.js'][23] = 0;
  _$jscoverage['util/zeropad.js'][24] = 0;
  _$jscoverage['util/zeropad.js'][25] = 0;
  _$jscoverage['util/zeropad.js'][26] = 0;
  _$jscoverage['util/zeropad.js'][27] = 0;
  _$jscoverage['util/zeropad.js'][28] = 0;
  _$jscoverage['util/zeropad.js'][29] = 0;
  _$jscoverage['util/zeropad.js'][31] = 0;
  _$jscoverage['util/zeropad.js'][32] = 0;
  _$jscoverage['util/zeropad.js'][33] = 0;
  _$jscoverage['util/zeropad.js'][34] = 0;
  _$jscoverage['util/zeropad.js'][35] = 0;
  _$jscoverage['util/zeropad.js'][36] = 0;
  _$jscoverage['util/zeropad.js'][37] = 0;
  _$jscoverage['util/zeropad.js'][38] = 0;
  _$jscoverage['util/zeropad.js'][39] = 0;
  _$jscoverage['util/zeropad.js'][40] = 0;
  _$jscoverage['util/zeropad.js'][41] = 0;
  _$jscoverage['util/zeropad.js'][42] = 0;
  _$jscoverage['util/zeropad.js'][43] = 0;
  _$jscoverage['util/zeropad.js'][44] = 0;
  _$jscoverage['util/zeropad.js'][47] = 0;
  _$jscoverage['util/zeropad.js'][48] = 0;
  _$jscoverage['util/zeropad.js'][49] = 0;
  _$jscoverage['util/zeropad.js'][50] = 0;
  _$jscoverage['util/zeropad.js'][53] = 0;
  _$jscoverage['util/zeropad.js'][54] = 0;
  _$jscoverage['util/zeropad.js'][55] = 0;
  _$jscoverage['util/zeropad.js'][56] = 0;
  _$jscoverage['util/zeropad.js'][58] = 0;
  _$jscoverage['util/zeropad.js'][61] = 0;
  _$jscoverage['util/zeropad.js'][62] = 0;
  _$jscoverage['util/zeropad.js'][63] = 0;
  _$jscoverage['util/zeropad.js'][64] = 0;
  _$jscoverage['util/zeropad.js'][65] = 0;
  _$jscoverage['util/zeropad.js'][66] = 0;
  _$jscoverage['util/zeropad.js'][67] = 0;
  _$jscoverage['util/zeropad.js'][68] = 0;
  _$jscoverage['util/zeropad.js'][69] = 0;
  _$jscoverage['util/zeropad.js'][70] = 0;
  _$jscoverage['util/zeropad.js'][71] = 0;
  _$jscoverage['util/zeropad.js'][72] = 0;
  _$jscoverage['util/zeropad.js'][73] = 0;
  _$jscoverage['util/zeropad.js'][74] = 0;
  _$jscoverage['util/zeropad.js'][75] = 0;
  _$jscoverage['util/zeropad.js'][78] = 0;
  _$jscoverage['util/zeropad.js'][79] = 0;
  _$jscoverage['util/zeropad.js'][80] = 0;
  _$jscoverage['util/zeropad.js'][81] = 0;
  _$jscoverage['util/zeropad.js'][82] = 0;
  _$jscoverage['util/zeropad.js'][83] = 0;
  _$jscoverage['util/zeropad.js'][84] = 0;
  _$jscoverage['util/zeropad.js'][85] = 0;
  _$jscoverage['util/zeropad.js'][88] = 0;
  _$jscoverage['util/zeropad.js'][89] = 0;
  _$jscoverage['util/zeropad.js'][90] = 0;
  _$jscoverage['util/zeropad.js'][91] = 0;
  _$jscoverage['util/zeropad.js'][92] = 0;
  _$jscoverage['util/zeropad.js'][95] = 0;
  _$jscoverage['util/zeropad.js'][96] = 0;
  _$jscoverage['util/zeropad.js'][97] = 0;
}
_$jscoverage['util/zeropad.js'][14]++;
"use strict";
_$jscoverage['util/zeropad.js'][17]++;
var INT_10_1 = Math.pow(10, 1);
_$jscoverage['util/zeropad.js'][18]++;
var INT_10_2 = Math.pow(10, 2);
_$jscoverage['util/zeropad.js'][19]++;
var INT_10_3 = Math.pow(10, 3);
_$jscoverage['util/zeropad.js'][20]++;
var INT_10_4 = Math.pow(10, 4);
_$jscoverage['util/zeropad.js'][21]++;
var INT_10_5 = Math.pow(10, 5);
_$jscoverage['util/zeropad.js'][22]++;
var INT_10_6 = Math.pow(10, 6);
_$jscoverage['util/zeropad.js'][23]++;
var INT_10_7 = Math.pow(10, 7);
_$jscoverage['util/zeropad.js'][24]++;
var INT_10_8 = Math.pow(10, 8);
_$jscoverage['util/zeropad.js'][25]++;
var INT_10_9 = Math.pow(10, 9);
_$jscoverage['util/zeropad.js'][26]++;
var INT_10_10 = Math.pow(10, 10);
_$jscoverage['util/zeropad.js'][27]++;
var INT_10_11 = Math.pow(10, 11);
_$jscoverage['util/zeropad.js'][28]++;
var INT_10_12 = Math.pow(10, 12);
_$jscoverage['util/zeropad.js'][29]++;
var INT_10_13 = Math.pow(10, 13);
_$jscoverage['util/zeropad.js'][31]++;
var ZERO_0 = "";
_$jscoverage['util/zeropad.js'][32]++;
var ZERO_1 = "0";
_$jscoverage['util/zeropad.js'][33]++;
var ZERO_2 = "00";
_$jscoverage['util/zeropad.js'][34]++;
var ZERO_3 = "000";
_$jscoverage['util/zeropad.js'][35]++;
var ZERO_4 = "0000";
_$jscoverage['util/zeropad.js'][36]++;
var ZERO_5 = "00000";
_$jscoverage['util/zeropad.js'][37]++;
var ZERO_6 = "000000";
_$jscoverage['util/zeropad.js'][38]++;
var ZERO_7 = "0000000";
_$jscoverage['util/zeropad.js'][39]++;
var ZERO_8 = "00000000";
_$jscoverage['util/zeropad.js'][40]++;
var ZERO_9 = "000000000";
_$jscoverage['util/zeropad.js'][41]++;
var ZERO_10 = "0000000000";
_$jscoverage['util/zeropad.js'][42]++;
var ZERO_11 = "00000000000";
_$jscoverage['util/zeropad.js'][43]++;
var ZERO_12 = "000000000000";
_$jscoverage['util/zeropad.js'][44]++;
var ZERO_13 = "0000000000000";
_$jscoverage['util/zeropad.js'][47]++;
var MAX_DECIMAL_LENGTH = 35;
_$jscoverage['util/zeropad.js'][48]++;
var ZEROS = exports.ZEROS = [""];
_$jscoverage['util/zeropad.js'][49]++;
for (var i = 1; i < MAX_DECIMAL_LENGTH; i++) {
  _$jscoverage['util/zeropad.js'][50]++;
  ZEROS.push(ZEROS[i - 1] + "0");
}
_$jscoverage['util/zeropad.js'][53]++;
exports.lpad = (function lpad(length, n) {
  _$jscoverage['util/zeropad.js'][54]++;
  var z = "" + n;
  _$jscoverage['util/zeropad.js'][55]++;
  while (z.length < length) {
    _$jscoverage['util/zeropad.js'][56]++;
    z = ZEROS[Math.min(length - z.length, MAX_DECIMAL_LENGTH - 1)] + z;
}
  _$jscoverage['util/zeropad.js'][58]++;
  return z;
});
_$jscoverage['util/zeropad.js'][61]++;
exports.lpad14 = (function lpad14(n) {
  _$jscoverage['util/zeropad.js'][62]++;
  if (n >= INT_10_13) {
    _$jscoverage['util/zeropad.js'][62]++;
    return ZERO_0 + n;
  }
  _$jscoverage['util/zeropad.js'][63]++;
  if (n >= INT_10_12) {
    _$jscoverage['util/zeropad.js'][63]++;
    return ZERO_1 + n;
  }
  _$jscoverage['util/zeropad.js'][64]++;
  if (n >= INT_10_11) {
    _$jscoverage['util/zeropad.js'][64]++;
    return ZERO_2 + n;
  }
  _$jscoverage['util/zeropad.js'][65]++;
  if (n >= INT_10_10) {
    _$jscoverage['util/zeropad.js'][65]++;
    return ZERO_3 + n;
  }
  _$jscoverage['util/zeropad.js'][66]++;
  if (n >= INT_10_9) {
    _$jscoverage['util/zeropad.js'][66]++;
    return ZERO_4 + n;
  }
  _$jscoverage['util/zeropad.js'][67]++;
  if (n >= INT_10_8) {
    _$jscoverage['util/zeropad.js'][67]++;
    return ZERO_5 + n;
  }
  _$jscoverage['util/zeropad.js'][68]++;
  if (n >= INT_10_7) {
    _$jscoverage['util/zeropad.js'][68]++;
    return ZERO_6 + n;
  }
  _$jscoverage['util/zeropad.js'][69]++;
  if (n >= INT_10_6) {
    _$jscoverage['util/zeropad.js'][69]++;
    return ZERO_7 + n;
  }
  _$jscoverage['util/zeropad.js'][70]++;
  if (n >= INT_10_5) {
    _$jscoverage['util/zeropad.js'][70]++;
    return ZERO_8 + n;
  }
  _$jscoverage['util/zeropad.js'][71]++;
  if (n >= INT_10_4) {
    _$jscoverage['util/zeropad.js'][71]++;
    return ZERO_9 + n;
  }
  _$jscoverage['util/zeropad.js'][72]++;
  if (n >= INT_10_3) {
    _$jscoverage['util/zeropad.js'][72]++;
    return ZERO_10 + n;
  }
  _$jscoverage['util/zeropad.js'][73]++;
  if (n >= INT_10_2) {
    _$jscoverage['util/zeropad.js'][73]++;
    return ZERO_11 + n;
  }
  _$jscoverage['util/zeropad.js'][74]++;
  if (n >= INT_10_1) {
    _$jscoverage['util/zeropad.js'][74]++;
    return ZERO_12 + n;
  }
  _$jscoverage['util/zeropad.js'][75]++;
  return ZERO_13 + n;
});
_$jscoverage['util/zeropad.js'][78]++;
exports.lpad7 = (function lpad7(n) {
  _$jscoverage['util/zeropad.js'][79]++;
  if (n >= INT_10_6) {
    _$jscoverage['util/zeropad.js'][79]++;
    return ZERO_0 + n;
  }
  _$jscoverage['util/zeropad.js'][80]++;
  if (n >= INT_10_5) {
    _$jscoverage['util/zeropad.js'][80]++;
    return ZERO_1 + n;
  }
  _$jscoverage['util/zeropad.js'][81]++;
  if (n >= INT_10_4) {
    _$jscoverage['util/zeropad.js'][81]++;
    return ZERO_2 + n;
  }
  _$jscoverage['util/zeropad.js'][82]++;
  if (n >= INT_10_3) {
    _$jscoverage['util/zeropad.js'][82]++;
    return ZERO_3 + n;
  }
  _$jscoverage['util/zeropad.js'][83]++;
  if (n >= INT_10_2) {
    _$jscoverage['util/zeropad.js'][83]++;
    return ZERO_4 + n;
  }
  _$jscoverage['util/zeropad.js'][84]++;
  if (n >= INT_10_1) {
    _$jscoverage['util/zeropad.js'][84]++;
    return ZERO_5 + n;
  }
  _$jscoverage['util/zeropad.js'][85]++;
  return ZERO_6 + n;
});
_$jscoverage['util/zeropad.js'][88]++;
exports.lpad4 = (function lpad4(n) {
  _$jscoverage['util/zeropad.js'][89]++;
  if (n >= INT_10_3) {
    _$jscoverage['util/zeropad.js'][89]++;
    return ZERO_0 + n;
  }
  _$jscoverage['util/zeropad.js'][90]++;
  if (n >= INT_10_2) {
    _$jscoverage['util/zeropad.js'][90]++;
    return ZERO_1 + n;
  }
  _$jscoverage['util/zeropad.js'][91]++;
  if (n >= INT_10_1) {
    _$jscoverage['util/zeropad.js'][91]++;
    return ZERO_2 + n;
  }
  _$jscoverage['util/zeropad.js'][92]++;
  return ZERO_3 + n;
});
_$jscoverage['util/zeropad.js'][95]++;
exports.lpad2 = (function lpad2(n) {
  _$jscoverage['util/zeropad.js'][96]++;
  if (n >= INT_10_1) {
    _$jscoverage['util/zeropad.js'][96]++;
    return ZERO_0 + n;
  }
  _$jscoverage['util/zeropad.js'][97]++;
  return ZERO_1 + n;
});
_$jscoverage['util/zeropad.js'].source = ["// Copyright 2013 SAP AG.","//","// Licensed under the Apache License, Version 2.0 (the \"License\");","// you may not use this file except in compliance with the License.","// You may obtain a copy of the License at","//","// http: //www.apache.org/licenses/LICENSE-2.0","//","// Unless required by applicable law or agreed to in writing,","// software distributed under the License is distributed on an","// \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,","// either express or implied. See the License for the specific","// language governing permissions and limitations under the License.","'use strict';","/* jshint curly: false */","","var INT_10_1 = Math.pow(10, 1);","var INT_10_2 = Math.pow(10, 2);","var INT_10_3 = Math.pow(10, 3);","var INT_10_4 = Math.pow(10, 4);","var INT_10_5 = Math.pow(10, 5);","var INT_10_6 = Math.pow(10, 6);","var INT_10_7 = Math.pow(10, 7);","var INT_10_8 = Math.pow(10, 8);","var INT_10_9 = Math.pow(10, 9);","var INT_10_10 = Math.pow(10, 10);","var INT_10_11 = Math.pow(10, 11);","var INT_10_12 = Math.pow(10, 12);","var INT_10_13 = Math.pow(10, 13);","","var ZERO_0 = '';","var ZERO_1 = '0';","var ZERO_2 = '00';","var ZERO_3 = '000';","var ZERO_4 = '0000';","var ZERO_5 = '00000';","var ZERO_6 = '000000';","var ZERO_7 = '0000000';","var ZERO_8 = '00000000';","var ZERO_9 = '000000000';","var ZERO_10 = '0000000000';","var ZERO_11 = '00000000000';","var ZERO_12 = '000000000000';","var ZERO_13 = '0000000000000';","","/* Decimal zero padding */","var MAX_DECIMAL_LENGTH = 35;","var ZEROS = exports.ZEROS = [''];","for (var i = 1; i &lt; MAX_DECIMAL_LENGTH; i++) {","  ZEROS.push(ZEROS[i - 1] + '0');","}","","exports.lpad = function lpad(length, n) {","  var z = '' + n;","  while (z.length &lt; length) {","    z = ZEROS[Math.min(length - z.length, MAX_DECIMAL_LENGTH - 1)] + z;","  }","  return z;","};","","exports.lpad14 = function lpad14(n) {","  if (n &gt;= INT_10_13) return ZERO_0 + n;","  if (n &gt;= INT_10_12) return ZERO_1 + n;","  if (n &gt;= INT_10_11) return ZERO_2 + n;","  if (n &gt;= INT_10_10) return ZERO_3 + n;","  if (n &gt;= INT_10_9) return ZERO_4 + n;","  if (n &gt;= INT_10_8) return ZERO_5 + n;","  if (n &gt;= INT_10_7) return ZERO_6 + n;","  if (n &gt;= INT_10_6) return ZERO_7 + n;","  if (n &gt;= INT_10_5) return ZERO_8 + n;","  if (n &gt;= INT_10_4) return ZERO_9 + n;","  if (n &gt;= INT_10_3) return ZERO_10 + n;","  if (n &gt;= INT_10_2) return ZERO_11 + n;","  if (n &gt;= INT_10_1) return ZERO_12 + n;","  return ZERO_13 + n;","};","","exports.lpad7 = function lpad7(n) {","  if (n &gt;= INT_10_6) return ZERO_0 + n;","  if (n &gt;= INT_10_5) return ZERO_1 + n;","  if (n &gt;= INT_10_4) return ZERO_2 + n;","  if (n &gt;= INT_10_3) return ZERO_3 + n;","  if (n &gt;= INT_10_2) return ZERO_4 + n;","  if (n &gt;= INT_10_1) return ZERO_5 + n;","  return ZERO_6 + n;","};","","exports.lpad4 = function lpad4(n) {","  if (n &gt;= INT_10_3) return ZERO_0 + n;","  if (n &gt;= INT_10_2) return ZERO_1 + n;","  if (n &gt;= INT_10_1) return ZERO_2 + n;","  return ZERO_3 + n;","};","","exports.lpad2 = function lpad2(n) {","  if (n &gt;= INT_10_1) return ZERO_0 + n;","  return ZERO_1 + n;","};"];
