// syracuse-ui/lib/common/tests/list/editListTest.js
"use strict";
var helpers = require("syracuse-core/lib/helpers")
var FieldsTest = require("../fieldsTest").FieldsTest;

function EditListTest(){
}

exports.EditListTest = helpers.defineClass(EditListTest, FieldsTest, {
    loadBox: function(){
        var $prototype = {
            "$": {
                "application/x-collection": {
                    "$title": "application/x-collection",
                    "$capabilities": "sort,filter,insert,append,delete,reorder",
                    "$type": "application/x-collection",
                    "$itemsPerPage": 5,
                    "$item": {
                        "$descriptor": "prototype employee.$edit",
                        "$type": "application/json",
                        "$pluralType": "children",
                        "$representation": "employee",
                        "$": {
                            "application/x-string": {
                                "$title": "x-string",
                                "$type": "application/x-string",
                                "$constraints": {
                                    "$maxLength": 30
                                },
                                "$capabilities": "sort,alphaTab"
                            },
                            "application/x-choice": {
                                $title: "application/x-choice",
                                $type: "application/x-choice",
                                $value: {
                                    $type: "application/x-integer",
                                    $constraints: {
                                        $enum: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10].map(function(value){
                                            return {
                                                "$value": value,
                                                $title: "Value " + value
                                            };
                                        })
                                    }
                                }
                            },
                            "application/x-date": {
                                $title: "x-date",
                                $type: "application/x-date"
                            },
                            "application/x-boolean": {
                                $title: "application/x-boolean",
                                $type: "application/x-boolean"
                            },
                            "application/x-decimal": {
                                $title: "application/x-decimal",
                                $type: "application/x-decimal",
                                $constraints: {
                                    "$isNullable": true
                                }
                            }
                        },
                        "$links": {},
                        "$actions": {
                            "$delete": {
                                "$title": "Supprimer",
                                "$method": "PUT"
                            }
                        }
                    },
                    "$actions": {
                        "$create": {
                            "$title": "Ajouter",
                            "$method": "PUT"
                        }
                    }
                }
            
            }
        };
        this._appendCapabilities($prototype);
        FieldsTest.prototype.loadBox.call(this, $prototype, {
            "application/x-collection": [{
                "$uuid": "166a1d1e-2a80-456a-a54f-1f62338aa443",
                "$etag": 1,
                "$": {},
                "$links": {
                    linkRecord1: {
                        $title: "linkRecord1"
                    },
                    linkRecord2: {
                        $title: "linkRecord2"
                    }
                },
                "$actions": {
                    actionRecord1: {
                        $title: "actionRecord1"
                    },
                    actionRecord2: {
                        $title: "actionRecord2"
                    }
                },
                "$index": 0,
                "application/x-string": "record 1",
                "application/x-choice": 2,
                "application/x-date": "2011-06-07",
                "application/x-boolean": true,
                "application/x-decimal": 2987111111111.4548,
                "parent": null
            }, {
                "$uuid": "df3f7fa4-7b64-4daa-b1f0-21352f39003f",
                "$etag": 1,
                "$": {},
                "$index": 1,
                "application/x-string": "record 2",
                "application/x-choice": 2,
                "application/x-date": "2011-06-07",
                "application/x-boolean": true,
                "application/x-decimal": 2987111111111.4548,
                "parent": null
            }, {
                "$uuid": "6369a2f6-3bb7-4663-9b9b-313d5b464110",
                "$etag": 1,
                "$": {},
                "$index": 2,
                "application/x-string": "record 3",
                "application/x-choice": 2,
                "application/x-date": "2011-06-07",
                "application/x-boolean": true,
                "application/x-decimal": 2987111111111.4548,
                "parent": null
            }],
            "$field$capabilityInsert": true,
            "$field$capabilityDelete": true,
            "$field$capabilityAppend": true,
            "$field$capabilityReorder": true
        });
    },
    _makeBlock: function($isEditMode){
        return [this._makeEditReadBlock("application/x-collection", [{
            $bind: "application/x-collection"
        }, {
            $bind: "application/x-collection",
            $layout: {
                $items: [{
                    $bind: "application/x-string"
                }]
            },
            "$cardviewPosition": "right",
            $cardview: {
                $layout: {
                    $items: [{
                        $layoutType: "columns",
                        $layoutSubType: "50,50",
                        $items: [{
                            $layoutType: "stack",
                            $items: [{
                                $bind: "application/x-string"
                            }]
                        }, {
                            $layoutType: "stack",
                            $items: [{
                                $bind: "application/x-choice"
                            }]
                        }]
                    }, {
                        $category: "block",
                        $isBoxCollapsable: true,
                        $opened: false,
                        $title: "test sub-block",
                        $layout: {
                            $layoutType: "tabs",
                            $items: [{
                                $category: "block",
                                $title: "tab1",
                                $layout: {
                                    $items: [{
                                        $bind: "application/x-date"
                                    }]
                                }
                            
                            }, {
                                $category: "block",
                                $title: "tab2",
                                $layout: {
                                    $items: [{
                                        $bind: "application/x-boolean"
                                    }, {
                                        $bind: "application/x-decimal"
                                    }]
                                }
                            
                            }]
                        }
                    }]
                
                }
            
            }
        }, {
            $bind: "application/x-collection",
            $format: "cardList",
            $cardview: {
                $layout: {
                    $items: [{
                        $layoutType: "columns",
                        $layoutSubType: "50,50",
                        $items: [{
                            $layoutType: "stack",
                            $items: [{
                                $bind: "application/x-string"
                            }]
                        }, {
                            $layoutType: "stack",
                            $items: [{
                                $bind: "application/x-choice"
                            }]
                        }]
                    }, {
                        $layoutType: "tabs",
                        $items: [{
                            $category: "block",
                            $title: "tab1",
                            $layout: {
                                $items: [{
                                    $bind: "application/x-date"
                                }]
                            }
                        
                        }, {
                            $category: "block",
                            $title: "tab2",
                            $layout: {
                                $items: [{
                                    $bind: "application/x-boolean"
                                }, {
                                    $bind: "application/x-decimal"
                                }]
                            }
                        
                        }]
                    }]
                
                }
            
            }
        }])];
    },
    appendSettingsBar: function(){
        return {
            $layoutType: "tabs",
            $items: [this._appendCapabilities(), FieldsTest.prototype.appendSettingsBar.call(this)]
        };
    }
});
