// syracuse-ui/lib/common/tests/list/fillListTest.js
"use strict";
var helpers = require("syracuse-core/lib/helpers")
var resources = require("../resources");
var Article = require("syracuse-ui/lib/common/article/article").Article;
function FillListTest(){
}

exports.FillListTest = helpers.defineClass(FillListTest, Article, {
    loadBox: function(initData){
        var self = this;
        self.$layoutOptions = document.site.$item.$layoutOptions.page["vignette"];
        self.$prototype = {
            "$properties": {
				"$field$type": {
                    $isHidden: true
                },
                "$field$isTopLabelAlignment": {
                    $isHidden: true
                },
                "$field$isRightTextLabelAlignment": {
                    $isHidden: true
                },
                $fieldData: {
                    $title: "records",
                    "$type": "application/x-string"
                },
                "application/x-collection": {
                    "$title": "application/x-collection",
                    "$capabilities": "sort,filter,insert,append,delete,reorder",
                    "$type": "application/x-collection",
                    "$itemsPerPage": 5,
                    "$item": {
                        "$descriptor": "prototype employee.$edit",
                        "$type": "application/json",
                        "$pluralType": "children",
                        "$representation": "employee",
                        "$properties": {
                            "application/x-string": {
                                "$title": "x-string",
                                "$type": "application/x-string",
                                
                                    "$maxLength": 30
                                ,
                                "$capabilities": "sort,alphaTab"
                            },
                            "application/x-choice": {
                                $title: "application/x-choice",
                                $type: "application/x-choice",
                                $value: {
                                    $type: "application/x-integer",
                                        $enum: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10].map(function(value){
                                            return {
                                                "$value": value,
                                                $title: "Value " + value
                                            };
                                        })
                                }
                            },
                            "application/x-date": {
                                $title: "x-date",
                                $type: "application/x-date"
                            },
                            "application/x-boolean": {
                                $title: "application/x-boolean",
                                $type: "application/x-boolean"
                            },
                            "application/x-decimal": {
                                $title: "application/x-decimal",
                                $type: "application/x-decimal",
                                
                                    "$isNullable": true
                                }
                        },
                        "$links": {},
                        "$actions": {
                            "$delete": {
                                "$title": "Supprimer",
                                "$method": "PUT"
                            }
                        }
                    },
                    "$actions": {
                        "$create": {
                            "$title": "Ajouter",
                            "$method": "PUT"
                        }
                    }
                }
            
            
            },
            "$actions": {
                "$empty": {
                    "$title": "Empty list",
                    "$method": "PUT"
                },
                "$refresh": {
                    "$title": "Refres list",
                    "$method": "PUT"
                }
            },
            "application/x-collection": resources.getRecords(8),
            $fieldData: JSON.stringify(resources.getRecords(8))
        };
        self.$item = {
            $isSeparatorsVisible: true,
            $layout: {
                $layoutType: "columns",
                $layoutSubType: "50,50",
                $items: [{
                    $layoutType: "stack",
                    $items: ["grid", "vgrid", "cards"].map(function($format){
                        return {
                            $category: "section",
                            $title: "Format " + $format,
                            $layout: {
                                $layoutType: "columns",
                                $layoutSubType: "50,50",
                                $items: [{
                                    $layoutType: "stack",
                                    $items: [self.makeField($format)]
                                }]
                            }
                        };
                    })
                
                }, {
                    $layoutType: "stack",
                    $items: [{
                        $layoutType: "row",
                        $items: [{
                            $skin: "s-list-record-menus-link",
                            $category: "link",
                            $bind: "$empty"
                        }, {
                            $skin: "s-list-record-menus-link",
                            $category: "link",
                            $bind: "$refresh"
                        }]
                    
                    }, {
                        $bind: "$fieldData",
                        $isEditMode: true,
                        $rows: 10
                    }]
                }]
            }
        };
        Article.prototype.loadBox.call(self, initData);
    },
    onNotifyRecordChange: function(value, binding){
        switch (binding) {
            case "$actions":
                var records = value.$empty ? [] : JSON.parse(this.boundFields.$fieldData[0].getDataValue());
                this.applyChange({
                    "application/x-collection": records
                })
                break;
                
        }
        return null; //cancel notify
    },
   
    makeField: function($format){
        return {
            $format: $format,
            $bind: "application/x-collection",
            $selectMode: "single",
            $layout: {
                $items: [{
                    "$bind": "application/x-string"
                }, {
                    "$bind": "application/x-choice"
                }, {
                    "$bind": "application/x-date"
                }, {
                    "$bind": "application/x-boolean"
                }, {
                    "$bind": "application/x-decimal"
                }]
            }
        };
    }
    
});
