// syracuse-ui/lib/common/tests/list/selectionListTest.js
"use strict";
var helpers = require("syracuse-core/lib/helpers")
var FieldsTest = require("../fieldsTest").FieldsTest;
var Article = require("syracuse-ui/lib/common/article/article").Article;
function SelectionListTest(){
}

exports.SelectionListTest = helpers.defineClass(SelectionListTest, Article, {
    loadBox: function(initData){
        var self = this;
        self.$layoutOptions = document.site.$item.$layoutOptions.page["vignette"];
        self.$prototype = {
            "$": {
                "application/x-collection": {
                    "$title": "application/x-collection",
                    "$capabilities": "sort,filter,insert,append,delete,reorder",
                    "$type": "application/x-collection",
                    "$itemsPerPage": 5,
                    "$item": {
                        "$descriptor": "prototype employee.$edit",
                        "$type": "application/json",
                        "$pluralType": "children",
                        "$representation": "employee",
                        "$": {
                            "application/x-string": {
                                "$title": "x-string",
                                "$type": "application/x-string",
                                "$constraints": {
                                    "$maxLength": 30
                                },
                                "$capabilities": "sort,alphaTab"
                            },
                            "application/x-choice": {
                                $title: "application/x-choice",
                                $type: "application/x-choice",
                                $value: {
                                    $type: "application/x-integer",
                                    $constraints: {
                                        $enum: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10].map(function(value){
                                            return {
                                                "$value": value,
                                                $title: "Value " + value
                                            };
                                        })
                                    }
                                }
                            },
                            "application/x-date": {
                                $title: "x-date",
                                $type: "application/x-date"
                            },
                            "application/x-boolean": {
                                $title: "application/x-boolean",
                                $type: "application/x-boolean"
                            },
                            "application/x-decimal": {
                                $title: "application/x-decimal",
                                $type: "application/x-decimal",
                                $constraints: {
                                    "$isNullable": true
                                }
                            }
                        },
                        "$links": {},
                        "$actions": {
                            "$delete": {
                                "$title": "Supprimer",
                                "$method": "PUT"
                            }
                        }
                    },
                    "$actions": {
                        "$create": {
                            "$title": "Ajouter",
                            "$method": "PUT"
                        }
                    }
                }
            
            
            },
            "application/x-collection": [{
                "$uuid": "df3f7fa4-7b64-4daa-b1f0-21352f39003f",
                "$etag": 1,
                "$": {},
                "$index": 1,
                "application/x-string": "record 2",
                "application/x-choice": 2,
                "application/x-date": "2011-06-07",
                "application/x-boolean": true,
                "application/x-decimal": 2987111111111.4548,
                "parent": null
            }, {
                "$uuid": "6369a2f6-3bb7-4663-9b9b-313d5b464110",
                "$etag": 1,
                "$": {},
                "$index": 2,
                "application/x-string": "record 3",
                "application/x-choice": 2,
                "application/x-date": "2011-06-07",
                "application/x-boolean": true,
                "application/x-decimal": 2987111111111.4548,
                "parent": null
            }]
        };
        self.$item = {
            $isSeparatorsVisible: true,
            $layout: {
                $layoutType: "stack",
                $items: ["grid", "vgrid"].map(function($format){
                    //$items: ["vgrid"].map(function($format){
                    return {
                        $category: "section",
                        $title: "Format " + $format,
                        $layout: {
                            $layoutType: "columns",
                            $layoutSubType: "50,50",
                            $items: [{
                                $layoutType: "stack",
                                $items: ["multi", "single"].map(function($selectMode){
                                    return {
                                        $title: "$selectMode = " + $selectMode,
                                        $format: $format,
                                        $bind: "application/x-collection",
                                        $selectMode: $selectMode,
                                        $layout: {
                                            $items: [{
                                                "$bind": "application/x-string"
                                            }, {
                                                "$bind": "application/x-choice"
                                            }]
                                        }
                                    };
                                })
                            }, {
                                $layoutType: "stack",
                                $items: ["button", "row"].map(function($selectMode){
                                    return {
                                        $title: "$selectMode = " + $selectMode,
                                        $format: $format,
                                        $bind: "application/x-collection",
                                        $selectMode: $selectMode,
                                        $layout: {
                                            $items: [{
                                                "$bind": "application/x-string"
                                            }, {
                                                "$bind": "application/x-choice"
                                            }]
                                        }
                                    };
                                })
                            }]
                        }
                    };
                })
            }
        };
        Article.prototype.loadBox.call(self, initData);
    }
    
    
});
