// syracuse-ui/lib/tests/list/arrayFieldTest.js
"use strict";
var helpers = require("syracuse-core/lib/helpers")
var FieldsTest = require("../fieldsTest").FieldsTest;
var ListAuthoring = require("syracuse-ui/lib/field/array/authoring/listAuthoring").ListAuthoring;
function ArrayFieldTest(){
}

exports.ArrayFieldTest = helpers.defineClass(ArrayFieldTest, FieldsTest, {
    onFetchHandler: function(options){
        if (options && options.params) {
            delete options.field;
            this.sendServerData(options);
        }
    },
    notifyChange: function($bind, value){
        switch ($bind) {
            case "$fieldRecordCount":
                this.refreshList();
                break;
            case "xarray":
                var sendBag = this.ensureSendBag();
                if (this.dataset && this.dataset.$uuid) {
                    sendBag.$uuid = this.dataset.$uuid;
                }
                sendBag[$bind] = value;
                sendBag.$etag = sendBag.$etag || 1;
                this.sendServerData(sendBag);
                break;
            case "$actions":
                if (value.$empty) {
                    this.refreshList([]);
                }
                else {
                    if (value.$refresh) {
                        this.refreshList();
                    }
                    else {
                        FieldsTest.prototype.notifyChange.call(this, $bind, value);
                    }
                }
                break;
            default:
                FieldsTest.prototype.notifyChange.call(this, $bind, value);
                break;
        }
    },
    loadBox: function(){
        FieldsTest.prototype.loadBox.call(this, this._makeArrayPrototype(), {
            $fieldRecordCount: 8
        });
        this._showCapablities();
        this.designedField = this.boundFields["xarray"][0];
        
        this._appendAuthoring();
        this.refreshList();
        
        var page = this.getPage();
        page.dataset.$etag = 1;
    },
    _appendAuthoring: function(){
        (this.authoring = new ListAuthoring()).loadBox(this);
        this.authoring.designedTestFields = this.boundFields["xarray"];
        this.idMap.authoringTest.$$body.empty().append(this.authoring.layoutContent.$$container);
        this.authoring.bindFieldEvents(this.authoring.layoutContent.$$container);
    },
    _makeArrayPrototype: function(){
        return {
            "$properties": {
                "$field$editMode": {
                    $title: "$editMode",
                    $type: "application/x-choice",
                    $isHidden: true,
                    $value: {
                        $type: "application/x-integer",
                        $enum: [{
                            $value: "array",
                            $title: "array (default)"
                        }, {
                            $value: "record",
                            $title: "record"
                        }, {
                            $value: "cell",
                            $title: "cell"
                        }]
                    }
                },
                "$field$type": {
                    $isHidden: true
                },
                "$field$isTopLabelAlignment": {
                    $isHidden: true
                },
                "$field$isRightTextLabelAlignment": {
                    $isHidden: true
                },
                $fieldRecordCount: {
                    $title: "Record Count",
                    "$type": "application/x-integer"
                },
                "xarray": {
                    "$title": "xarray",
                    "$capabilities": "sort,filter,insert,append,delete,reorder",
                    "$type": "application/x-array",
                    "$itemsPerPage": 5,
                    "$item": {
                        "$descriptor": "prototype employee.$edit",
                        "$type": "application/json",
                        "$pluralType": "children",
                        "$representation": "employee",
                        "$properties": {
                            "xstring": {
                                "$title": "xstring",
                                "$type": "application/x-string",
                                
                                "$maxLength": 30,
                                "$capabilities": "sort,filter,alphaTab"
                            },
                            "xchoice": {
                                $title: "xchoice",
                                $type: "application/x-choice",
                                $value: {
                                    $type: "application/x-integer",
                                    $enum: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10].map(function(value){
                                        return {
                                            "$value": value,
                                            $title: "Value " + value
                                        };
                                    })
                                }
                            },
                            "xdate": {
                                $title: "xdate",
                                $type: "application/x-date",
                                "$capabilities": "sort,filter,alphaTab"
                            },
                            "xboolean": {
                                $title: "xboolean",
                                $type: "application/x-boolean",
                                "$capabilities": "sort,filter"
                            },
                            "xdecimal": {
                                $title: "xdecimal",
                                $type: "application/x-decimal",
                                "$capabilities": "sort,filter",
                                
                                "$isNullable": true
                            }
                        },
                        "$links": {},
                        "$actions": {}
                    },
                    "$actions": {}
                }
            },
            "$actions": {
            
                "$empty": {
                    "$title": "Empty list",
                    "$method": "PUT"
                },
                "$refresh": {
                    "$title": "Refresh list",
                    "$method": "PUT"
                }
            }
        };
    },
    refreshList: function(records){
        if (records === undefined) {
            var $fieldRecordCount = this.boundFields.$fieldRecordCount[0].getDataValue();
            this.boundFields.$field$dataset[0].setDataValue(JSON.stringify(records = this.getResources().getRecords($fieldRecordCount)));
        }
        this.applyChange({
            $properties: {
                "xarray": {
                    $totalResults: records.length
                }
            },
            $field$editMode: "array",
            "xarray": records
        });
    },
    
    appendSettingsBar: function(){
        var $items = FieldsTest.prototype.appendSettingsBar.call(this);
        $items.unshift({
            $layoutType: "stack",
            $items: [{
                $layoutType: "row",
                $items: [{
                    $skin: "s-list-record-menus-link",
                    $category: "link",
                    $bind: "$empty"
                }, {
                    $skin: "s-list-record-menus-link",
                    $category: "link",
                    $bind: "$refresh"
                }]
            
            }, {
                $bind: "$fieldRecordCount",
                $isEditMode: true
            }, {
                $bind: "$field$editMode",
                $isEditMode: true
            }]
        });
        $items.push({
            $category: "block",
            $title: "Authoring",
            $id: "authoringTest",
            $css: "s-authoring-list-test",
            $layout: {
                $items: []
            }
        });
        return $items;
    }
});
