// syracuse-ui/lib/tests/list/cardsTest.js
"use strict";
var helpers = require("syracuse-core/lib/helpers")
var AuthoringListTest = require("./authoringListTest").AuthoringListTest;
function CardsTest(){
}

exports.CardsTest = helpers.defineClass(CardsTest, AuthoringListTest, {
    _getCard: function(){
        return {
            $bind: "application/x-array",
            $format: "cards",
            $cards: {
                $layout: {
                    $items: [{
                        $layoutType: "columns",
                        $layoutSubType: "50,50",
                        $items: [{
                            $layoutType: "stack",
                            $items: [{
                                $bind: "application/x-string"
                            }]
                        }, {
                            $layoutType: "stack",
                            $items: [{
                                $bind: "application/x-choice"
                            }]
                        }]
                    }, {
                        $layoutType: "tabs",
                        $items: [{
                            $category: "block",
                            $title: "tab1",
                            $layout: {
                                $items: [{
                                    $bind: "application/x-date"
                                }]
                            }
                        
                        }, {
                            $category: "block",
                            $title: "tab2",
                            $layout: {
                                $items: [{
                                    $bind: "application/x-boolean"
                                }, {
                                    $bind: "application/x-decimal"
                                }]
                            }
                        
                        }]
                    }]
                
                }
            
            }
        };
    },
    _makeBlock: function($isEditMode){
        return [this._makeEditReadBlock("application/x-array", [this._getCard()]), this._getCard()];
    }
    
});
