// syracuse-ui/lib/tests/list/deltaTests.js
exports.fullDelta = [{
    $title: "insert 1",
    records: [{
        "$uuid": "insert 1",
        "$index": 1,
        "xstring": "insert 1"
    }]
}, {
    $title: "insert 2",
    records: [{
        "$uuid": "insert 1",
        "$index": 1,
        "xstring": "insert 1"
    }, {
        "$uuid": "insert 5",
        "$index": 5,
        "xstring": "insert 5"
    }]
}, {
    $title: "upd 2 4",
    records: [{
        "$uuid": "record2",
        "$index": 1,
        "xstring": "updated 2",
        "$properties": {
            "xstring": {
                "$fieldStyle": "color:blue;"
            }
        }
    }, {
        "$uuid": "record 4",
        "$index": 3,
        "xstring": "updated 4",
        "$properties": {
            "xstring": {
                "$fieldStyle": "color:red;",
                "$diagnoses": [{
                    message: "test erreur"
                }]
            }
        }
    }, {
        "$uuid": "record7",
        "$index": 6,
        "xstring": "updated 7",
        "$properties": {
            "xstring": {
                "$fieldStyle": "font-size:1.9em;background-color:green;color:white;",
                "$diagnoses": [{
                    message: "test info",
                    severity: "info"
                }]
            }
        }
    }]
}];
exports.defaultDelta = [{
    $title: "insert 1",
    records: [{
        "$uuid": "insert 1",
        "$index": 1,
        "xstring": "insert 1"
    }]
}, {
    $title: "upadate record 2",
    records: [{
        "$uuid": "record2",
        "$index": 2,
        "xstring": "updated 2"
    }]
}, {
    $title: "upadate 2 3 6 7 ",
    records: [{
        "$uuid": "record0"
    }, {
        "$uuid": "record1"
    }, {
        "$uuid": "record2",
        "xstring": "updated 2"
    }, {
        "$uuid": "record3",
        "xstring": "updated 3"
    }, {
        "$uuid": "record4"
    }, {
        "$uuid": "record5"
    }, {
        "$uuid": "record6",
        "xstring": "updated 6"
    }, {
        "$uuid": "record7",
        "xstring": "updated 7"
    }]
}];
