// syracuse-ui/lib/tests/list/editListTest.js
"use strict";
var helpers = require("syracuse-core/lib/helpers")
var ArrayFieldTest = require("./arrayFieldTest").ArrayFieldTest;

function EditListTest(){
}

exports.EditListTest = helpers.defineClass(EditListTest, ArrayFieldTest, {
    _makeBlock: function($isEditMode){
        return [{
            $bind: "xarray",
            $isEditMode: true
        }, {
            $bind: "xarray",
            $format: "vgrid",
            $isEditMode: true
        }, {
            $bind: "xarray",
            $isEditMode: true,
            $format: "cards",
            $cardsByRowCount: 2,
            $cards: {
                $layout: {
                    $items: [{
                        $layoutType: "columns",
                        $layoutSubType: "50,50",
                        $items: [{
                            $layoutType: "stack",
                            $items: [{
                                $bind: "xstring"
                            }]
                        }, {
                            $layoutType: "stack",
                            $items: [{
                                $bind: "xchoice"
                            }]
                        }]
                    }, {
                        $layoutType: "tabs",
                        $items: [{
                            $category: "block",
                            $title: "tab1",
                            $layout: {
                                $items: [{
                                    $bind: "xdate"
                                }]
                            }
                        
                        }, {
                            $category: "block",
                            $title: "tab2",
                            $layout: {
                                $items: [{
                                    $bind: "xboolean"
                                }, {
                                    $bind: "xdecimal"
                                }]
                            }
                        
                        }]
                    }]
                
                }
            
            }
        }];
    },
    
    appendSettingsBar: function(){
        var $items = ArrayFieldTest.prototype.appendSettingsBar.call(this);
        $items.push({
            $category: "block",
            $title: "$actions",
            $layout: {
                $items: [{
                    $category: "link",
                    $bind: "$deleteItem",
                    $isEditMode: true
                }, {
                    $category: "link",
                    $bind: "$insertItem",
                    $isEditMode: true
                }, {
                    $category: "link",
                    $bind: "$reorderItem",
                    $isEditMode: true
                }]
            }
        });
        return $items;
    }
});
