// syracuse-ui/lib/tests/list/formatArrayTest.js
"use strict";
var helpers = require("syracuse-core/lib/helpers")
var ArrayFieldTest = require("./arrayFieldTest").ArrayFieldTest;

function GridTest(){
}

exports.GridTest = helpers.defineClass(GridTest, ArrayFieldTest, {
    _makeBlock: function($isEditMode){
        return [this._makeEditReadBlock("xarray", [{
            $bind: "xarray"
        }]), {
            $bind: "xarray"
        }];
    }
});

function ScrollTest(){
}

exports.ScrollTest = helpers.defineClass(ScrollTest, ArrayFieldTest, {
    _makeBlock: function($isEditMode){
        return [this._makeEditReadBlock("xarray", [{
            $bind: "xarray"
        }]), {
            $bind: "xarray"
        }];
    }
});

function VgridTest(){
}

exports.VgridTest = helpers.defineClass(VgridTest, ArrayFieldTest, {
    _makeBlock: function($isEditMode){
        return [this._makeEditReadBlock("xarray", [{
            $bind: "xarray",
            $format: "vgrid",
        }]), {
            $bind: "xarray",
            $format: "vgrid",
        }];
    }
});



function CardsTest(){
}

exports.CardsTest = helpers.defineClass(CardsTest, ArrayFieldTest, {
    _getCard: function(){
        return {
            $bind: "xarray",
            $format: "cards",
            $cards: {
                $layout: {
                    $items: [{
                        $layoutType: "columns",
                        $layoutSubType: "50,50",
                        $items: [{
                            $layoutType: "stack",
                            $items: [{
                                $bind: "xstring"
                            }]
                        }, {
                            $layoutType: "stack",
                            $items: [{
                                $bind: "xchoice"
                            }]
                        }]
                    }, {
                        $layoutType: "tabs",
                        $items: [{
                            $category: "block",
                            $title: "tab1",
                            $layout: {
                                $items: [{
                                    $bind: "xdate"
                                }]
                            }
                        
                        }, {
                            $category: "block",
                            $title: "tab2",
                            $layout: {
                                $items: [{
                                    $bind: "xboolean"
                                }, {
                                    $bind: "xdecimal"
                                }]
                            }
                        
                        }]
                    }]
                
                }
            
            }
        };
    },
    _makeBlock: function($isEditMode){
        return [this._makeEditReadBlock("xarray", [this._getCard()]), this._getCard()];
    }
});

function FillListTest(){
}

exports.FillListTest = helpers.defineClass(FillListTest, ArrayFieldTest, {
    _makeBlock: function($isEditMode){
        return ["grid", "vgrid", "cards"].map(function($format){
            return {
                $format: $format,
                $bind: "xarray",
                $selectMode: "single",
                $cardsByRowCount: 3,
                $layout: {
                    $items: [{
                        "$bind": "xstring"
                    }, {
                        "$bind": "xchoice"
                    }, {
                        "$bind": "xdate"
                    }, {
                        "$bind": "xboolean"
                    }, {
                        "$bind": "xdecimal"
                    }]
                }
            };
        });
    },
    refreshList: function(records){
        ArrayFieldTest.prototype.refreshList.call(this, records);
        this.applyChange({
            $properties: {
                "$field$editMode": {
                    $isHidden: false
                }
            }
        });
    }
});

function SelectionListTest(){
}

exports.SelectionListTest = helpers.defineClass(SelectionListTest, ArrayFieldTest, {
    _makeBlock: function($isEditMode){
        var self = this;
        return ["grid", "vgrid", "cards"].map(function($format){
            return {
                $category: "section",
                $title: "Format " + $format,
                $layout: {
                    $layoutType: "columns",
                    $layoutSubType: "50,50",
                    $items: [{
                        $layoutType: "stack",
                        $items: ["multi", "single"].map(function($selectMode){
                            return self.makeField($selectMode, $format);
                        })
                    }, {
                        $layoutType: "stack",
                        $items: ["button", "row"].map(function($selectMode){
                            return self.makeField($selectMode, $format);
                        })
                    }]
                }
            };
        });
    },
    makeField: function($selectMode, $format){
        return {
            $title: "$selectMode = " + $selectMode,
            $format: $format,
            $bind: "xarray",
            $selectMode: $selectMode,
            $cardsByRowCount: 2,
            $layout: {
                $items: [{
                    "$bind": "xstring"
                }, {
                    "$bind": "xchoice"
                }]
            }
        };
    }
    
});
