// @sage/syracuse-react/src/navigation/sitemap/page/__tests__/testSetup.ts
"use strict";
const sc = require('./syra_core.json');

const BODY_ID = "s_app_body";
const SLOT_ID = "s_app_slot";
process.env.NODE_ENV = 'test';

module.exports = {
    setUp: () => {
        global.sessionStorage = {
            getItem: function (key: any) {
                return this[key];
            },
            setItem: function (key: any, value: any) {
                this[key] = value;
            },
            removeItem: function (key: any) {
                delete this[key];
            }
        };

        global.localStorage = {
            getItem: function (key: any) {
                return this[key];
            },
            setItem: function (key: any, value: any) {
                this[key] = value;
            },
            removeItem: function (key: any) {
                delete this[key];
            }
        };
        global.syra_core = sc;
        global.syra_menus = {
            click: {
                fire: jest.fn()
            }
        }
        let el = document.createElement('div');
        el.setAttribute('id', 's_app_slot');
        document.body.appendChild(el);
    }
}
