// syracuse-tablet/html/js/pages/drafts/actionAdapterPageDrafts.js
"use strict";

var Base = require('syracuse-tablet/html/js/pages/actionAdapter').ActionAdapter;
var utils = require('syracuse-tablet/html/js/helpers/utils');
var locale = require('syracuse-tablet/html/js/helpers/locale');

var _headersLinks = [{
	action: "draftsRunStop",
	title: locale.text("drafts.action.stop"),
	hidden: true
}, {
	action: "draftsRunSynch",
	title: locale.text("drafts.action.synch"),
	disabled: true
}, {
	action: "draftsRemove",
	title: locale.text("drafts.action.remove"),
	disabled: true
}];
/**
 * 
 */
exports.ActionAdapterPageDrafts = utils.defineClass(
	function ActionAdapterPageDrafts(page) {
		Base.call(this, page);
	},
	Base, {
		destroy: function() {
			Base.prototype.destroy.call(this);
		},
		getHeaderItems: function() {
			var items = $.extend(true, [], _headersLinks);
			var self = this;
			items.forEach(function(item) {
				item.icon = item.action;
				item["controller-id"] = self.page.controller.id;
			})
			return items;
		},
		getActionsPanelItems: function() {
			var items = [];
			return items;
		}
	});
