// syracuse-tablet/html/js/helpers/articleParser.js
"use strict";

var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("articleParser");
var utils = require('syracuse-tablet/html/js/helpers/utils');
var ctrlFactory = require('syracuse-tablet/html/js/controls/ctrlFactory');
/**
 * Set default array $article if no $article found
 * @Mathias:  It's the inconvenient to share a same JSON object for all elements
 *			  We need to set default $ article in advanced (here) instead of control/layout class
 */

var _SINGLEARRAYPROP = "~data~";
var _isSingleArray = function(itemProto) {
	return itemProto.$properties == null || itemProto.$type == "application/x-reference";
};
var _setArrayDefAuthoring = function(itemProto, article) {
	var self = this;
	article.$article = {
		$display: "table",
		$items: []
	};
	var props;
	if (_isSingleArray(itemProto)) {
		props = {};
		props[_SINGLEARRAYPROP] = itemProto;
	} else {
		props = itemProto.$properties;
	}
	// TODO - Optimization - itemProto will be recreated by array
	var prop;
	for (var propName in props) {
		if (!propName.smStartsWith("$")) {
			prop = props[propName];
			if (prop.$isExcluded !== true) {
				article.$article.$items.push({
					"$bind": propName
				});
			}
		}
	}
	return article;
};
/**
 * Create an article form fieldProto
 * Accept a JSON or prototype class object
 */
var _createArticle = function(fieldName, fieldProto, options) {
	var a = $.extend({}, {
		"$bind": fieldName
	}, options);
	var isJson = $.isPlainObject(fieldProto),
		singleArray;
	if (isJson ? fieldProto.$type === "application/x-array" : fieldProto.isArray()) {
		_setArrayDefAuthoring(isJson ? fieldProto.$item : fieldProto.data("$item"), a);
	}
	return a;
};
/**
 * Parses an article and generate the control tree structure
 */
var _Parser = utils.defineClass(
	function(page) {
		// Controller
		var self = this;
		self.page = page;
	}, null, {
		doParse: function(article, prototype, parent, layoutPath, unfounds) {
			var self = this;
			var num = 0;
			if (article.$layoutType) {
				var lyt = ctrlFactory.createLayout(self.page, parent, article, {
					layoutPath: layoutPath
				});
				if (lyt && article.$items) {
					if (prototype.isQuery()) {
						prototype = prototype.getPrototype("$resources");
					}
					article.$items.forEach(function($itm) {
						// Important: Do not clone article here since full tree has to be preserved	
						self.doParse($itm, prototype, lyt, layoutPath + "-" + num, unfounds);
						num++;
					});
				}
				return lyt;
			} else if (article.$bind) {
				var fieldProto = prototype.getPrototype(article.$bind);
				if (fieldProto == null && article.$isUnfound !== true) {
					unfounds.push(article);
				} else {
					if (article.$isUnfound) {
						fieldProto = prototype.create({
							$type: "tablet/x-unfound"
						});
					}
					if (!fieldProto.$isExcluded) {
						if (fieldProto.isArray() && article.$article == null) {
							// Add default authoring
							_setArrayDefAuthoring(fieldProto, article);
						}
						// Important: Do not clone article here since full tree has to be preserved
						ctrlFactory.createControl(self.page, parent, article, fieldProto, {
							layoutPath: layoutPath
						});
					}
				}
			} else {
				log && log("$layoutType or $bind exspected");
				//throw new Error ("$layoutType or $bind expected");
			}
		},
		parse: function(article, prototype, opts) {
			var self = this;
			try {
				if (!article || !prototype) throw new Error("Not empty article and prototype are expected");
				if (!article.$layoutType) throw new Error("$layoutType expected for root layout");
				var unfounds = [];
				var root = self.doParse(article, prototype, null, (opts && opts.layoutPath) || "0", unfounds);
				if (unfounds.length > 0) {
					var article = {
						$layoutType: "stack",
						$items: []
					};
					unfounds.forEach(function(f) {
						article.$items.push($.extend(true, f, {
							$isUnfound: true
						}));
					});
					self.doParse(article, prototype, root, (opts && opts.layoutPath) || "0");
				}
				return root;
			} catch (e) {
				log && log('article parser error', e);
				throw new Error("Error parsing article [" + e.message + "]\n" + utils.cleanStack(e.stack));
			}
		}
	}
);

/**
 * Parses article and calls the controls factory
 * 		article: json description
 *  	prototype: prototype object (Prototype class)
 *  Return root element (layout)
 */
var _article2Controls = function(page, article, prototype, opts) {
	var p = new _Parser(page);
	return p.parse(article, prototype, opts);
};


var _cardV2Controls = function(arrayCtrl, options) {
	var article = arrayCtrl.article.$article.$card;
	if (!article) {
		article = {
			"$layoutType": "stack",
			"$items": []
		};
		arrayCtrl.article.$article.$items.forEach(function(item) {
			article.$items.push(item);
		});
	}
	options = options || {};
	options.isArrayChild = true;
	options.isTableCell = false;
	return _article2Controls(arrayCtrl.controller, article, arrayCtrl.prototype.getPrototype("$item"), options);
};
/**
 * Convert a desktop article to tablet one
 * 		article:  desktop json description
 *  	prototype: prototype object (Prototype class)
 */
var _defOption = {
	skipblock: true,
	skipsection: true,
	include: ["$category", "$layoutType", "$title", "$bind"],
	exclude: ["$items", "$widths"],
	protoExclude: ["application/x-array", "application/json"]
};

var _convertWidth = function(widths) {
	var r = [];
	widths = widths.split(",");
	// temporarily - We can also embed cells into 
	if (widths.length === 0) throw new Error("Number of cells/rows must be > 0");
	if (widths.length > 12) throw new Error("Number of cells/rows must be <= 12");
	var sum = 0;
	widths.forEach(function(w) {
		w = parseInt(w, 10);
		w = Math.ceil(w / 12);
		sum += w;
		r.push(w);
	});
	if (sum > 12) {
		var d = sum - 12;
		r.sort();
		while (d > 0) {
			// Remove excess starting from large cells
			for (var i = r.length - 1; i >= 0; i--) {
				if (d > 0 && r[i] > 1) {
					r[i] = r[i] - 1;
					d--;
				}
			}
		}
	} else if (sum < 12) {
		var d = 12 - sum;
		r.sort();
		while (d > 0) {
			// Add shortage starting from small cells
			for (var i = 0; i < r.length; i++) {
				if (d > 0) {
					r[i] = r[i] + 1;
					d--;
				}
			}
		}
	}
	return r.join(",");
};
var _scanDesktop = function(json, options) {

	if (!json) return null;
	// we skip $layout
	var res;
	if (json.$layout) {
		res = _scanDesktop(json.$layout, options);
	} else {
		res = {};
	}
	if (json.$layoutType) {
		if (!json.$items) throw new Error("$items is missing");
		res.$items = [];
		json.$items.forEach(function(itm) {
			res.$items.push(_scanDesktop(itm, options));
		});
	}
	for (var p in json) {
		if (p === "$widths") {
			res[p] = _convertWidth(json[p]);
		} else if (options.exclude.indexOf(p) === -1 && options.include.indexOf(p) >= 0) {
			res[p] = json[p];
		}
	}
	return res;
};

var _convertDesktop = function(article, prototype, options) {
	options = $.extend({}, _defOption, options);
	return _scanDesktop(article, options);
};

exports.cardV2Controls = _cardV2Controls;
exports.article2Controls = _article2Controls;
exports.convertDesktop = _convertDesktop;
exports.createArticle = _createArticle;
