// syracuse-tablet/html/js/helpers/dashboardUtils.js
"use strict";

var prototype = require('syracuse-tablet/html/js/helpers/prototype');
var authArticleGen = require('syracuse-tablet/html/js/authoring/authoringArticleGen');

/**
 *  Generates proto form dashBoard description
 **/
var _getDashboardProto = function(dash) {
	var proto = {
		$title: dash.$title,
		$description: dash.$description,
		$properties: {},
		$dashboardName: dash.$dashboardName
	};
	var vignetteIds = Object.keys(dash.$vignettes || []);
	for (var i = 0; i < vignetteIds.length; i++) {
		var vignetteId = vignetteIds[i];
		var property = $.extend({}, dash.$vignettes[vignetteId]);
		var gadgetId = property.$uuid;

		property.$gadget = dash.$gadgets[gadgetId];
		if (property.$gadget) {
			property.$gadget = $.extend({}, property.$gadget);
			// Set gadget uuid
			property.$gadget.$uuid = gadgetId;
			// Set vignette uuid
			property.$uuid = vignetteId;
			// Set type - "application/x-string"
			property.$type = "tablet/x-vignette";
			property.$title = property.$gadget.$title;
		} else {
			property.$gadget = {
				$uuid: gadgetId,
				$type: "$gadgetMissing"
			};
			property.$uuid = vignetteId;
			property.$type = "tablet/x-vignette";

		}
		proto.$properties[vignetteId] = property;
	}
	return prototype.create(proto);
};

/**
 * Returns $article from dashBoard info
 * 		Calculates and Sets $article if no $article was found
 */
var _getDashboardArticle = function(proto) {
	return authArticleGen.generateDashboardArticle(proto, "orientation");
};

var _getDashboardInfo = function(dashboardMetaData) {
	return $.smResolve()
		.then(function() {
			var article = dashboardMetaData.$article;
			var proto = _getDashboardProto(dashboardMetaData);
			if (!article) {
				return _getDashboardArticle(proto)
					.then(function(article) {
						dashboardMetaData.$article = article;
						return {
							article: $.extend({}, article),
							prototype: proto
						};
					});
			} else {
				return {
					article: $.extend({}, article),
					prototype: proto
				};
			}
		});

};

exports.getDashboardInfo = _getDashboardInfo;
exports.getDashboardArticle = _getDashboardArticle;
exports.getDashboardProto = _getDashboardProto;
