// syracuse-tablet/html/js/helpers/formController.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var reqProto = require('syracuse-tablet/html/js/helpers/prototype');
var Base = require('syracuse-tablet/html/js/helpers/controller').Klass;

var _createProto = function(protoJson, protoParent, baseUrl) {
	if (!protoParent && baseUrl) {
		protoJson.$baseUrl = baseUrl;
	}
	return reqProto.create(protoJson, protoParent);
};

var _createArticle = function(proto) {
	var article = {
		"$layoutType": "stack",
		"$items": []
	};
	var protoJson = proto.data("$properties");
	Object.keys(protoJson).forEach(function(key) {
		if (!protoJson[key].$isExcluded) {
			article.$items.push({
				"$bind": key
			});
		}
	});
	return article;
};
/**
 * options
 * 		$activity	read by default
 */
var _Klass = utils.defineClass(
	function(id, proto, article, dao, options) {
		if (article == null) {
			article = _createArticle(proto);
		}
		Base.call(this, id, proto, article, options);
		this.setActivity(this.options.$activity);
		this.setDao(dao);
	}, Base, {
		destroy: function() {
			Base.prototype.destroy.call(this);
		},
		render: function($$parent, opts) {
			if (this.rootLayout == null) {
				this.buildControlsTree();
			}
			var opts = $.extend(opts, {
				afterRender: true
			});
			Base.prototype.render.call(this, $$parent, opts);
		},
		/**
		 * Show element in page - scroll if needed
		 * c is a JQuery elmt or an object with $$elmt property
		 */
		makeVisible: function(c) {
			// TODO - Implement a gestureManager in base controller to allow makeVisible
		}
	});


exports.createProto = _createProto;
exports.Klass = _Klass;
