// syracuse-tablet/html/js/helpers/locale.js
"use strict";

var globals = require('syracuse-tablet/html/js/helpers/globals');
var ajax = require('syracuse-tablet/html/js/ajax/ajax').getAjax().ajax;

var _currentLocale = null;
var _resources = {};

var _datetimeInformationKey = "$datetimeInformation";
var _decimalInformationKey = "$decimalInformation";
var _upLocalePreferences;

var _txtRegexp = /\{([\w-]+)\}/g;
var _localeHashUser = "*";

var _userProfileSettings = {};
var _userLocales = {};

var _text = function(key, args) {
	var resources = _resources && _resources[_currentLocale];
	if (!resources) {
		return;
	}
	var text = resources[key] || "Resource key not found[" + key + "]";

	if (args) {
		text = text.replace(_txtRegexp, function(match, p1) {
			var idx = +p1;
			if (idx >= 0 && idx < args.length) {
				return args[idx];
			}
			return "?";
		});
	}

	return text;
};

var _getDatetimeInfo = function() {
	var resources = _resources && _resources[_currentLocale];
	if (!resources) {
		return;
	}
	return resources[_datetimeInformationKey];
};

var _getDecimalInfo = function() {
	var resources = _resources && _resources[_currentLocale];
	if (!resources) {
		return;
	}
	return resources[_decimalInformationKey];
};

var _setLocale = function(locale, upLocalePreferences) {
	var deferred = new $.Deferred();
	_upLocalePreferences = upLocalePreferences;

	// Everytime a locale is set, the user is remembered
	// this is because formatters created afterwards are cached
	// and need to be linked to locale + user (because of user defined locale settings)
	// See getCurrentLocaleHash()
	var ctx = globals.getUserCtx();
	_localeHashUser = ctx && ctx.$user || "*";
	_userProfileSettings = _userLocales[locale] || {};

	var resources = _resources && _resources[locale];
	if (resources) {
		_currentLocale = locale;
		deferred.resolve();
	} else if (locale === _currentLocale) {
		deferred.resolve();
	} else {
		var result = ajax("GET", "/syracuse-tablet/html/js/resources/strings-" + locale + ".json", null, null);
		_loadFile("/syracuse-tablet/html/js/resources", "strings", locale, "en")
			.then(function(data) {
				_currentLocale = locale;
				_resources[locale] = data;
			})
			.then(function() {
				return _loadFile("/syracuse-tablet/html/js/resources/locales", "date", locale, "en-GB");
			})
			.then(function(data) {
				// join strings-xx-XX and date-xx-XX
				_resources[locale][_datetimeInformationKey] = data;
			})
			.then(function() {
				return _loadFile("/syracuse-tablet/html/js/resources/locales", "decimal", locale, "en-GB");
			})
			.then(function(data) {
				// join strings-xx-XX and decimal-xx-XX
				_resources[locale][_decimalInformationKey] = data;
			})
			.then(function() {
				deferred.resolve();
			}, function(e) {
				deferred.resolve();
			});
	}
	return deferred.promise();
};

var _loadFile = function(path, file, locale, defLocale) {
	var deferred = new $.Deferred();
	var segs = locale.split("-");

	function _load() {
		if (segs.length === 0) {
			// Fallback to default
			var url = path + "/" + file + "-" + defLocale + ".json";
			ajax("GET", url, null, null)
				.then(function(data) {
					if (data && data.responseJSON) {
						deferred.resolve(data.responseJSON);
					}
					deferred.resolve({});
				}, function(e) {
					deferred.reject(e);
				});
		} else {
			var code = segs.join("-");
			var url = path + "/" + file + "-" + code + ".json";
			ajax("GET", url, null, null)
				.then(function(data) {
					if (data && data.responseJSON) {
						deferred.resolve(data.responseJSON);
					} else if (data.status === 404) {
						segs.pop();
						_load();
					} else {
						deferred.reject(data.textStatus + " - " + data.status);
					}
				}, function(e) {
					deferred.reject(e);
				});
		}
	}
	_load();
	return deferred.promise();
};

var _isCurrent = function(locale, strict) {
	if (!locale || !_currentLocale) return false;
	if (_currentLocale === locale) return true;
	return strict ? false : (locale.split('-'))[0] === _currentLocale;
};

/*
 * TODO: Use prefered format from _upLocalePreferences if set
 */
exports.getDateFormat = function() {
	var di = _getDatetimeInfo();
	return _userProfileSettings.shortDate || (di && di.formatPatterns && di.formatPatterns.shortDate) || "dd/MM/yyyy";
};

exports.getDateTimeFormat = function() {
	var di = _getDatetimeInfo();
	var date = _userProfileSettings.shortDate || (di && di.formatPatterns && di.formatPatterns.shortDate) || "dd/MM/yyyy";
	var time = _userProfileSettings.shortTime || (di && di.formatPatterns && di.formatPatterns.shortTime) || "HH:mm:ss";
	return date + " " + time;
};
exports.getTimeFormat = function() {
	var di = _getDatetimeInfo();
	return _userProfileSettings.longTime || (di && di.formatPatterns && di.formatPatterns.longTime || "HH:mm:ss");
};
exports.getTimeFormatShort = function() {
	var di = _getDatetimeInfo();
	return _userProfileSettings.shortTime || (di && di.formatPatterns && di.formatPatterns.shortTime || "HH:mm");
};

exports.getNumberFormat = function(type) {
	return (type === "application/x-integer" ? "#,##0" : "#,##0.##");
};

exports.getNumberGroupSeparator = function() {
	var di = _getDecimalInfo();
	var gs = _userProfileSettings.numberGroupSeparator || (di && di.numberGroupSeparator) || ",";
	return gs;
};

exports.getNumberDecimalSeparator = function() {
	var di = _getDecimalInfo();
	var ds = _userProfileSettings.numberDecimalSeparator || (di && di.numberDecimalSeparator) || ".";
	return ds;
};

exports.setLocale = _setLocale;
exports.text = _text;
exports.isCurrent = _isCurrent;

// Get Month names, etc for parsers
exports.getDatetimeInfo = _getDatetimeInfo;
exports.getCurrentLocale = function() {
	return _currentLocale;
};
exports.getCurrentLocaleHash = function() {
	return _currentLocale + "#" + _localeHashUser;
};
exports.setUserLocales = function(locales) {
	_userLocales = {};
	locales.forEach(function(locale) {
		_userLocales[locale.code] = {
			shortDate: locale.shortDate,
			shortTime: locale.shortTime,
			longTime: locale.longTime,
			numberDecimalSeparator: locale.numberDecimalSeparator,
			numberGroupSeparator: locale.numberGroupSeparator
		};
	});
};
