// syracuse-tablet/html/js/helpers/mobileDetect.js
"use strict";

var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("mobileDetect", false);
log = null;
/**
 * Npm module that detects browsers - Used on desktop
 */
var _bowser = require('syracuse-tablet/html/deps/bowser/bowser');

var _defaultOS = "windows";

/**
 * Default initialization
 * http://hgoebl.github.io/mobile-detect.js/doc/MobileDetect.html#MobileDetect
 */
var _MobileDetect = (function() {
	return window.MobileDetect ? new window.MobileDetect(window.navigator.userAgent) : null;
})();

/**
 * called by application when user agent has changed (test device on chrome degugger)
 */
exports.init = function() {
	if (window.bowser && window.bowser._detect) {
		_bowser = window.bowser._detect(window.navigator.userAgent);
	}
	_MobileDetect = window.MobileDetect ? new window.MobileDetect(window.navigator.userAgent) : null;
	if (log) {
		log("User agent Name", window.navigator.userAgent);
		log("Browser Name: ", exports.getBrowserName());
		log("Browser OS: ", _MobileDetect.os() || "Not detected (desktop)");
		log && log("Browser OS Style: ", exports.getStyleOS());
	}
};
//Returns the os for style ios/windows/android
exports.getStyleOS = function() {
	var os = _MobileDetect == null ? _defaultOS : _MobileDetect.os();
	switch (os) {
		case "AndroidOS":
			return "android";
		case "WindowsMobileOS":
		case "WindowsPhoneOS":
			return "windows";
		case "iOS":
			return "ios";
		case "PalmOS":
		case "BlackBerryOS":
		case "MeeGoOS":
		case "MaemoOS":
		case "JavaOS":
		case "webOS":
		case "badaOS":
		case "BREWOS":
		case "SymbianOS":
		default:
			return _defaultOS;
	}
};
// check ios/windows/android
// http://hgoebl.github.io/mobile-detect.js/doc/MobileDetect.html#os
exports.isOs = function(os) {
	if (_MobileDetect == null) {
		return os === _defaultOS;
	}
	switch (_MobileDetect.os()) {
		case "AndroidOS":
			return os === "android";
		case "WindowsMobileOS":
		case "WindowsPhoneOS":
			return os === "windows";
		case "iOS":
			return os === "ios";
		case "PalmOS":
		case "BlackBerryOS":
		case "MeeGoOS":
		case "MaemoOS":
		case "JavaOS":
		case "webOS":
		case "badaOS":
		case "BREWOS":
		case "SymbianOS":
		default:
			return os === _defaultOS;
	}
};
// http://hgoebl.github.io/mobile-detect.js/doc/MobileDetect.html#mobile
exports.isMobile = function() {
	return _MobileDetect && _MobileDetect.mobile() !== null;
};
var _mobileDetectCheckBrowser = function(name, log) {
	if (_MobileDetect == null) return null;
	var res = _MobileDetect.userAgent();
	if (!res) return null;
	res = res.toLowerCase();
	res = (res === "ie" && name === "msie") || res === name;
	if (res === true) {
		log && log(name, "detect by mobile-detect");
	}
	return res;
};
var _bowserCheckBrowser = function(name, log) {
	if (_bowser == null) return null;
	var res = _bowser[name] === true;
	if (res === true) {
		log && log(name, "detect by bowser");
	}
	return res;
};
var _checkBrowserName = function(name, log) {
	// use bowser for desktop and mobile-detect for mobile
	// FDB I did a lot of tests and it's the better way
	// - Eg safari on IOS is not detected as Safari by bowser but mobile-detect doesn't detect correctly browser on a desktop...
	var res = (exports.isMobile() ? _mobileDetectCheckBrowser : _bowserCheckBrowser)(name, log);
	if (res == null) {
		res = (exports.isMobile() ? _bowserCheckBrowser : _mobileDetectCheckBrowser)(name, log);
	}
	return res;
};
/** Key (case insensitive):
 * Accepts chrome, firefox, msie, msedge, opera, phantom, safari, seamonkey
 * Accepts multiple arguments msie, msedge...
 */
exports.checkBrowserName = function() {
	var result = false,
		name, bn;
	for (var i = 0; i < arguments.length; i++) {
		result = result || _checkBrowserName(arguments[i].toLowerCase());
	}
	return result;
};
/**
 * Just for info
 */
exports.getBrowserName = function() {
	var res = exports.isMobile() ? _MobileDetect.userAgent() : _bowser.name;
	if (res == null) {
		res = exports.isMobile() ? _bowser.name : _MobileDetect.userAgent();
	}
	return res ? res.toLowerCase() : null;
};

exports.testBrowserName = function(log) {
	var result = {};
	["chrome", "firefox", "msie", "msedge", "opera", "phantom", "safari", "seamonkey"].forEach(function(b) {
		result[b] = _checkBrowserName(b, log);
	});
	//log && log(JSON.stringify(_bowser, null, 2));
	log && log(JSON.stringify(result, null, 2));
	return result;
};

exports.init();
