// syracuse-tablet/html/js/helpers/modals.js
"use strict";

var log = require('syracuse-tablet/html/js/helpers/logger').log;
var utils = require('syracuse-tablet/html/js/helpers/utils');
var templates = require('syracuse-tablet/html/js/helpers/templating');


var _error = function(title, e, cb) {
	if (typeof e === "object" && e.fileName && e.stack) {
		e = utils.toDiagnose({
			where: title,
			exception: e
		});
	} else if (typeof e === "string") {
		e = utils.toDiagnose({
			where: title,
			message: e
		});
	} else {
		e = utils.toDiagnose(e);
	}
	var body = JSON.stringify(e, null, 2);
	log && log("modals.displayError", body);
	_modal("error", {
		close: "Close",
		title: title || "Error",
		body: body
	}, cb);
};
var _confirm = function(action, cb) {
	_modal("confirm", {
		title: "Confirm",
		action: action,
		yes: "yes",
		no: "no"
	}, function(action) {
		if (cb) cb("yes" === action);
	});
};
var _modal = function(id, context, cb) {
	id = "modals_" + id;
	$.when(templates.execSync(id, context)).always(function(html) {
		if (html == null) {
			alert("Modal not found - id=" + id + "\n" + JSON.stringify(context, null, 2));
			if (cb) cb("close");
		} else {
			var h = $(html).modal();
			var done = false;
			// Use <button data-dismiss"modal" for close action
			// Use <button data-action"myaction" for action other than close 
			h.on("hidden.bs.modal", function() {
				// Called when dialog is close - data-dismiss = "modal" or  h.modal('hide')
				h.unbind();
				h.remove();
				// Call cb only if no click action
				if (!done && cb) cb("close");
			});
			// For dialogs with option
			h.click("[data-action]", function(evt) {
				var t = $(evt.target);
				var act = t.attr("data-action");
				if (act) {
					done = true;
					evt.preventDefault();
					evt.stopPropagation();
					// Close modal
					h.modal('hide');
					if (cb) cb(act, t.attr("data-params"));
				}
			});
		}
	});
};

var _init = function() {
	templates.preload(["modals/error", "modals/confirm"]);
};
exports.init = _init;
exports.modal = _modal;
exports.error = _error;
exports.confirm = _confirm;
