// syracuse-tablet/html/js/helpers/offlineManager.js
"use strict";

var locale = require('syracuse-tablet/html/js/helpers/locale');
var globals = require('syracuse-tablet/html/js/helpers/globals');;
var notifications = require('syracuse-tablet/html/js/helpers/notifications');

var _OffLineMgr = function() {

	this._offline = false;

	this.setOffline = function(offline, cause) {
		offline = offline === true;
		if (offline === this._offline) return;
		this._offline = offline;
		globals.getModal().notify({
			severityClass: offline ? "warning" : "success",
			title: locale.text("info.network." + (offline ? "offline" : "online") + ".title"),
			body: locale.text("info.network." + (offline ? "offline.outdated" : "online.uptodate"))
		});
		$("#s-m-app-id").toggleClass("offline", offline);
		notifications.publish("sm.offline.state.change", offline); // not used for moment ....
	};

	this.isOnline = function() {
		return navigator.onLine && this._offline !== true;
	};
};


exports.init = function() {
	globals.setOfflineManager(new _OffLineMgr());
};
