// syracuse-tablet/html/js/helpers/offlineManager.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var globals = require('syracuse-tablet/html/js/helpers/globals');;
var notifications = require('syracuse-tablet/html/js/helpers/notifications');

var _Klass = utils.defineClass(
	function() {
		this._offline = false;
	}, null, {
		destroy: function() {

		},
		// Be careful "notif" parameter not null or not undefined, we force notification display!
		setOffline: function(offline, notif) {
			offline = offline || false;
			var stateChange = offline !== this._offline;
			this._offline = offline;
			if (stateChange || notif) {
				notif = notif || {};
				notif = $.extend(true, {
					severityClass: offline ? "warning" : "success",
					title: locale.text("info.network." + (offline ? "offline" : "online") + ".title"),
					body: locale.text("info.network." + (offline ? "offline.outdated" : "online.uptodate"))
				}, notif);
				globals.getModal().notify(notif);
				$("#s-m-app-id").toggleClass("offline", offline);
			}
		},
		isOffline: function() {
			return this._offline;
		}

	});
exports.Klass = _Klass;

exports.create = function() {
	return new _Klass();
};
