// syracuse-tablet/html/js/helpers/offlineManager.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var locale = require('syracuse-tablet/html/js/helpers/locale');
var globals = require('syracuse-tablet/html/js/helpers/globals');;
var notifications = require('syracuse-tablet/html/js/helpers/notifications');

var _Klass = utils.defineClass(
	function() {
		this._offline = false;
	}, null, {
		destroy: function() {

		},
		setOffline: function(offline) {
			offline = offline || false;
			if (offline !== this._offline) {
				this._offline = offline;
				globals.getModal().notify({
					severityClass: offline ? "warning" : "success",
					title: locale.text("info.network." + (offline ? "offline" : "online") + ".title"),
					body: locale.text("info.network." + (offline ? "offline.outdated" : "online.uptodate"))
				});
				$("#s-m-app-id").toggleClass("offline", offline);
				notifications.publish("sm.offline.state.change", offline); // not used for moment ....
			}
		},
		isOffline: function() {
			return this._offline;
		}

	});
exports.Klass = _Klass;

exports.create = function() {
	return new _Klass();
};
