// syracuse-tablet/html/js/helpers/siteLayout.js
"use strict";


/*
 *
 * This module is responsibe to layout main layout components in the browser window:
 *
 * HEADER    | AUTHORING
 * CONTENT   | PANEL
 * FOOTER    |
 *
 */

var globals = require('syracuse-tablet/html/js/helpers/globals');
var notifications = require('syracuse-tablet/html/js/helpers/notifications');

var _currentOrientation = null;
var _layoutParameters = {
	header: {
		height: 50
	},
	footer: {
		height: 50
	},
	authPanelRight: {
		width: 300
	}
};

/*
 * To detect orientation change in browser window compared to last call of this function
 */
function _isOrientationChange() {
	var width = $(window).width();
	var height = $(window).height();

	var ratio = width / height;
	var orientation;

	if (ratio < 1) {
		orientation = "portrait";
	} else {
		orientation = "landscape";
	}
	if (orientation == _currentOrientation) {
		return false;
	}
	_currentOrientation = orientation;
	return true;
}

/*
 * Triggered by browser window change
 */
exports.updateLayout = function(evt, opts) {
	// If browser window size change but orientation did not change, the the software keyboard has been rendered and
	// there MUST NOT be a resize to not loose focus
	if (!_isOrientationChange()) {
		return;
	}

	_layoutMainContainers();
};

/*
 * Triggered by application JS code
 */
exports.updateLayoutInternal = function(evt, opts) {
	if (opts && opts.layout) {
		_layoutParameters = $.extend(true, _layoutParameters, opts.layout);
	}
	_layoutMainContainers();
};

/*
 * Layout main containers like header, content, footer, authoring panel
 */
function _layoutMainContainers() {
	var width = $(window).width();
	var height = $(window).height();

	var authPanelRightWidth = 0;
	var $$auth = $("#s-m-auth-id");
	if ($$auth.is(":visible")) {
		authPanelRightWidth = _layoutParameters.authPanelRight.width;
		_layoutAuthPanelRight($$auth, width, height);
	}

	var applicationWidth = width - authPanelRightWidth;
	var applicationHeight = height;

	var $$app = $("#s-m-app-id");

	var $$header = $(".s-m-page.s-m-full > header");
	var $$content = $(".s-m-page.s-m-full > section.s-m-main-content");
	var $$footer = $(".s-m-page.s-m-full > footer");

	var contentWidth = applicationWidth;
	var contentHeight = applicationHeight - _layoutParameters.header.height - _layoutParameters.footer.height;

	// Container including header, content, footer
	$$app.width(applicationWidth);
	$$app.height(applicationHeight);

	// Position header, content, footer
	$$header.css({
		top: 0,
		left: 0,
		width: contentWidth,
		height: _layoutParameters.header.height
	});
	$$content.css({
		top: _layoutParameters.header.height,
		left: 0,
		width: contentWidth,
		height: contentHeight
	});
	$$footer.css({
		top: _layoutParameters.header.height + contentHeight,
		left: 0,
		width: contentWidth,
		height: _layoutParameters.header.height
	});

	// Allow other components like authoring panel or page content to be resized
	notifications.publish("sm.main.layout.changed");
}

function _layoutAuthPanelRight($$auth, width, height) {
	$$auth.css({
		left: width - _layoutParameters.authPanelRight.width,
		top: 0,
		width: _layoutParameters.authPanelRight.width,
		height: height
	});
}
