// syracuse-tablet/html/js/helpers/templating.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var log = require('syracuse-tablet/html/js/helpers/logger').getLogger("tmpl");
var globals = require('syracuse-tablet/html/js/helpers/globals');
var ajax = require('syracuse-tablet/html/js/ajax/ajax').getAjax().ajax;

var _config = {
	htmlHolder: "body",
	tmplPath: "templates",
	tmplExt: ".tmpl"
};
var _temptates = {};

var _url = (function() {
	var url = window.location.href.split('/');
	url[url.length - 1] = _config.tmplPath + "/";
	return url.join('/');
})();

var _get = function(filePath, tmplId, options) {
	options = options || {};
	var deferred = $.Deferred();
	tmplId = tmplId || filePath;
	var domId = utils.domId(tmplId);
	if (_temptates[tmplId]) {
		deferred.resolve(_temptates[tmplId]);
	} else if ($("#" + domId).length > 0) {
		_temptates[tmplId] = $("#" + domId).html();
		deferred.resolve(_temptates[tmplId]);
	} else {
		var url = _url + filePath + _config.tmplExt;
		if (globals.isDvlpMode()) {
			// Force templae reload in development mode
			url += "?" + new Date().getTime();
		}
		ajax("GET", url, null, {
			"Accept": "text/html"
		}).always(function(result) {
			try {
				var data = result.responseText || "";
				if (result.status === 200) {
					// We add the script tag here to have Syntax highlighting in .tmpl file
					var e = $(_config.htmlHolder);
					if (data.indexOf("<script") === -1) {
						e.append($('<script id="' + domId + '" type="text/x-handlebars-template">' + data + "</script>"));
					} else {
						$(data).each(function(i, script) {
							e.append(script);
						});
					}
					deferred.resolve($("#" + domId).html());
				} else {
					if (options.failNotFound) throw new Error("Template not found status[" + result.status + "]");
					deferred.resolve(null);
				}
			} catch (e) {
				deferred.reject(utils.toDiagnose({
					exception: e,
					where: "stemplates._get [" + filePath + "]"
				}));
			}
		});
	}
	return deferred.promise();
};

// file : template file path - tmplId : template id - tmplId = null if only one template in the file
var _exec = function(filePath, context, options, tmplId) {
	var deferred = $.Deferred();
	tmplId = tmplId || filePath;
	_get(filePath, tmplId, options).then(function(tmpl) {
		try {
			if (typeof tmpl === "string") {
				tmpl = Handlebars.compile(tmpl);
				_temptates[tmplId] = tmpl;
			}
			deferred.resolve(tmpl ? tmpl(context || {}) : null);
		} catch (e) {
			deferred.reject({
				message: "Url = [" + _url + "]\n\tfilePath[#" + filePath + "]\n\ttmplId[#" + tmplId + "]\n\tExt [" + _config.tmplExt + "]",
				exception: e,
				where: "stemplates.exec"
			});
		}
	}, function(e) {
		deferred.reject(e);
	});
	return deferred.promise();
};

// Execute a template already loaded - Multiple templates in a same file
var _execSync = function(tmplId, context) {
	var tmpl = _temptates[tmplId];
	if (!_temptates[tmplId]) {
		tmpl = $("#" + utils.domId(tmplId)).html();
		if (tmpl) {
			tmpl = Handlebars.compile(tmpl);
			_temptates[tmplId] = tmpl;
		}
	}
	return tmpl ? tmpl(context || {}) : null;
};

// Execute a template already loaded - Multiple templates in a same file
var _execHtmlTmpl = function(html, context) {
	var tmpl = _temptates[tmplId];
	if (!_temptates[tmplId]) {
		tmpl = $("#" + utils.domId(tmplId)).html();
		if (tmpl) {
			tmpl = Handlebars.compile(tmpl);
			_temptates[tmplId] = tmpl;
		}
	}
	return tmpl ? tmpl(context || {}) : null;
};


// Preload templates in background - TODO use workers
var _preload = function(templates) {
	var promises = [];
	templates.forEach(function(t) {
		promises.push(_exec(t), {});
	});
	$.when(promises).always(function() {
		log && log("Templates preload", "Count[" + templates.length + "]", "loaded[" + arguments.length + "]");
	});
};

var _setConfig = function(config, rooturl) {
	if (rooturl) _url = rooturl + '/' + _config.tmplPath + '/';
};

//id = null -> Remove all templates
var _remove = function(id) {
	var s = id ? "#" + utils.domId(id) : "script[type='text/x-handlebars-template']";
	var e = $(_config.htmlHolder).find(s).remove();
};

/* called by main.js*/
var _init = function() {
	// Remove templates on reload otherwise they not refreshed
	if (globals.isDvlpMode()) {
		// Remove templates to force reload in development mode
		_remove();
	}
};
exports.setConfig = _setConfig;
exports.preload = _preload;
exports.remove = _remove;
exports.execHtmlTmpl = _execHtmlTmpl;
exports.execSync = _execSync;
exports.exec = _exec;
exports.init = _init;
