// syracuse-tablet/html/js/helpers/unitTestManager.js
"use strict";

var globals = require('syracuse-tablet/html/js/helpers/globals');
var _application = globals.getApplication();

var _notifPageLoaded = function(step, page) {
	if (step == "before") {
		$('[data-s-ut-page-is-current]').removeAttr("data-s-ut-page-is-current");
		return;
	}
	if (step == "after") {
		if (!page) {
			return;
		}
		// Put page name on vignettes too once they are loaded so we can easily find them too
		page.$$elmt.attr("data-s-ut-page-name", page.state.name);
		if (page.isVignette) {
			return;
		}
		// Put current page flag on top level page only
		page.$$elmt.attr("data-s-ut-page-is-current", page.state.name);
		return;
	}
};
var _notifUserCtxChanged = function(ctx, endpoint) {
	globals.getApplication().$$elmt.removeAttr("data-s-ut-user-ctx");
	if (ctx && endpoint) {
		var data = [ctx.$user || "", ctx.$role || "", ctx.$lang || "", endpoint || ""];
		globals.getApplication().$$elmt.attr("data-s-ut-user-ctx", data.join(','));
	}
};
var _notifDeviceChanged = function(device, orientation) {
	globals.getApplication().$$elmt.attr("data-s-ut-device", device + "," + orientation);
};
var _arrayAddRowInfo = function($$elmt, idx, rowData) {
	$$elmt.attr("data-s-ut-row-idx", idx);
};
var _addStructElmtId = function($$elmt, info) {
	$$elmt.attr("data-s-ut-elmt-id", info);
};
var _authAddRightPanelId = function($$elmt, id) {
	if (id) {
		$$elmt.attr("data-s-ut-auth-selected-elmt", id);
	} else {
		$$elmt.removeAttr("data-s-ut-auth-selected-elmt");
	}

};
var _authAddTreeElmtId = function($$elmt, info) {
	$$elmt.attr("data-s-ut-auth-elmt-id", info);
};

var _init = function() {
	globals.setUnitTestMgr(exports);
};

exports.notifPageLoaded = _notifPageLoaded;
exports.notifUserCtxChanged = _notifUserCtxChanged;
exports.notifDeviceChanged = _notifDeviceChanged;
exports.arrayAddRowIndex = _arrayAddRowInfo;
exports.addStructElmtId = _addStructElmtId;
exports.authAddRightPanelId = _authAddRightPanelId;
exports.authAddTreeElmtId = _authAddTreeElmtId;
exports.init = _init;
