// syracuse-tablet/html/js/helpers/userPrefs.js
"use strict";

var utils = require('syracuse-tablet/html/js/helpers/utils');
var localStorage = require('syracuse-tablet/html/js/storage/localStorage');
var globals = require('syracuse-tablet/html/js/helpers/globals');

var _Klass = utils.defineClass(function userPrefs(type, id) {
	this._id = id;
	this._type = type;
}, null, {
	destroy: function() {},
	_read: function() {
		var info = localStorage.getItem(this._getStorageId());
		if (!info) return null;
		try {
			return JSON.parse(info);
		} catch (e) {};
		return null;
	},
	_getStorageId: function() {
		return globals.getUserCtx().$user + "_" + this._id + "_" + this._type;
	},
	getPrefs: function() {
		return this._read(this) || {};
	},
	savePrefs: function(opts) {
		var prefs = null;
		if (opts !== null) {
			var prefs = this.getPrefs();
			for (var p in opts) {
				prefs[p] = opts[p];
			}
		}
		localStorage.setItem(this._getStorageId(), prefs ? JSON.stringify(prefs) : null);
	},
	reset: function(opts) {
		this.savePrefs(null);
	}
});

exports.createUserPrefs = function(type, id) {
	return new _Klass(type, id);
};
