// @sage/syracuse-ui/lib/field/array/tools/graph.js
"use strict";
var _slotResizer = require("@sage/syracuse-ui/lib/field/array/tools/slotResizer");

function _load(list, load) {
	// if (!list.chartField && (load || list.$item.$graphPosition != "behind")) {
	if (!list.chartField && (load || list.$item.$graphPosition != "behind")) {
		var $graphItem = syra_dataset.applyDelta(list.page, {}, list.$item);
		$graphItem.$isDecoratorType = true;
		$graphItem.$category = "field";
		$graphItem.attachedField = list;
		list.chartField = list.page.addItem(list.graphSlot, $graphItem, list.articleParent);
	}
}

function _draw(list) {
	if (list.$switchToList !== false) {
		if (list.$item.$graphPosition == "front" || list.$item.$graphPosition == "behind") {
			var icon, title;
			if (list.$item.$graphPosition == "front") {
				icon = "field_array";
				title = syra_local.flSwitchToList;
			} else {
				icon = "field_graph";
				title = syra_local.flSwitchToGraph;
			}
			if (!list.toggleGraphBtn) {
				list.toggleGraphBtn = syra_button.add({
					parent: list,
					text: title,
					css: list.$mnPickersCss,
					fontIcon: icon,
					click: function() {
						this.parent.toggleGraph();
					}
				});
				list.topbar.insertBefore(list.toggleGraphBtn.link, list.searcherSlot);
			} else {
				syra_button.setText(list.toggleGraphBtn, title, icon);
				syra_button.hide(list.toggleGraphPicker, false);
			}
		} else {
			syra_button.hide(list.toggleGraphBtn, true);
		}
		_load(list, true);
		if (!list.page.isFusionPage) {
			syra_menus.applyChange(list, {
				$links: {
					$pptslide: {
						$title: list.chartField.extendWidget.locale.highCharts_powerpointSlide,
						$url: list.page.$prototype.$representationUrl + "&count={recordsPerPage}" + "&pptMode=newSlide&$bind=" + list.$item.$bind + "&{linkstartP}",
						$type: "application/syracuse-ppt-slide",
						$confirm: list.chartField.extendWidget.locale.highCharts_installOfficeAddin,
						$officeAddinSetup: syra_officeAddin.exe + "?&format=application/x-msi"
					}
				}
			});
		}
	} else {
		list.$isQuickDesignerDisabled = true;
		list.$item.$graphPosition = "front";
		_load(list);
	}
	syra_dom.hide(list.builder.tableSlot, list.$item.$graphPosition == "front" && !list.builder.isOutCardMode);
	syra_dom.hide(list.graphSlot, list.$item.$graphPosition == "behind");
}


exports.isDefined = function(list) {
	return !!(list.$prototype.$cube || list.$prototype.$decorator);
};

exports.draw = function(list, load) {
	if (exports.isDefined(list)) {
		if (!list.toggleGraph) {
			list.toggleGraph = function() {
				this.$item.$graphPosition = this.$item.$graphPosition == "front" ? "behind" : "front";
				_draw(this);
				this.paging && this.paging.showPagers();
				this.resizeItem(true);
			};
			list.addGrapSlot = function() {
				var tableSlot = this.builder.tableSlot;
				if (!this.graphResizer) {
					this.graphResizer = _slotResizer.addGraphResizer(list);
				}
				if (!this.graphSlot) {
					this.graphSlot = document.createElement("div");
				}
				var separator = this.graphResizer.context.handle;
				switch (this.$item.$graphPosition) {
					case "left":
						tableSlot.parentNode.insertBefore(separator, tableSlot);
						tableSlot.parentNode.insertBefore(this.graphSlot, separator);
						separator.className = "s-array-graph-sep-cell";
						this.graphSlot.className = "s-array-graph-slot-cell";
						if (this.$item.$graphWidth) {
							this.graphSlot.style.width = this.$item.$graphWidth;
						}
						break;
					case "right":
						tableSlot.parentNode.insertBefore(separator, tableSlot.nextSibling);
						tableSlot.parentNode.appendChild(this.graphSlot, separator.nextSibling);
						separator.className = "s-array-graph-sep-cell";
						this.graphSlot.className = "s-array-graph-slot-cell";
						if (this.$item.$graphWidth) {
							this.graphSlot.style.width = this.$item.$graphWidth;
						}
						break;
					case "top":
						var sibling = this.tableRow || tableSlot;
						sibling.parentNode.insertBefore(separator, sibling);
						sibling.parentNode.insertBefore(this.graphSlot, separator);
						separator.className = "s-array-graph-sep";
						this.graphSlot.className = "s-array-graph-slot";
						this.graphSlot.style.width = "";
						break;
					case "front":
					case "behind":
					case "bottom":
						var sibling = this.tableRow || tableSlot;
						sibling.parentNode.appendChild(separator);
						sibling.parentNode.appendChild(this.graphSlot);
						separator.className = "s-array-graph-sep";
						this.graphSlot.className = "s-array-graph-slot";
						this.graphSlot.style.width = "";
						break;

				}
			};
			list.removeGrapSlot = function() {
				if (this.graphResizer) {
					syra_dom.remove(this.graphResizer.context.handle);
					this.graphResizer.dispose();
				}
				syra_dom.remove(this.graphSlot);
				this.graphResizer = this.graphSlot = null;
			};
			list.builder.list_onAfterSetDataValue = function(dataRecordSet, parentDataRecord, metaData) {
				var list = this.list;
				if (list.chartField && list.records) {
					var dataRecords = [];
					if (list.chartField.$useFullData) {
						dataRecords = list.dataset.slice(0, list.dataset.length);
					} else {
						for (var ii = 0, jj = list.records.length; ii < jj; ii++) {
							dataRecords.push(list.records[ii].dataset);
						}
					}
					if (list.page.isFusionPage && dataRecords.length > 0) {
						if (syra_fusion.sapUtil.isEmptyGridDataRecordEx(list.$fields, dataRecords)) {
							dataRecords.splice(dataRecords.length - 1);
						}
					}
					list.chartField.setValue(dataRecords, metaData, parentDataRecord);
				}
			};
			list.applyGrapDesignMetaData = function(metaData, designing) {
				if (metaData.$graphPosition !== undefined) {
					this.$item.$graphPosition = metaData.$graphPosition;
				}
				if (metaData.$cube) {
					this.$item.$cube = metaData.$cube;
				}
				designing && this.builder.appendCardGraphSlots();
				_draw(this);
				this.chartField && this.chartField.applyDesignMeta && this.chartField.applyDesignMeta(metaData);
			};

		}
		list.$item.$graphPosition = list.$item.$graphPosition || "bottom";
		load && _load(list, true);
	}
};
