// @sage/syracuse-lib/src/collaboration/updateScripts/updateScript_F_120636_office_document_upload.ts
"use strict";

exports.tracer; // = console.log;

var _scripts = [];

_scripts[1] = function(_, db) {
	exports.tracer && exports.tracer("Executing update script to version: 1; add option in security profile to allow upload of Office documents");
	//
	var coll = db.db.collection("SecurityProfile", _);
	coll.update({
		allowOfficeDocumentUpload: {
			$exists: false
		}
	}, {
		$set: {
			_updDate: new Date(),
			allowOfficeDocumentUpload: true
		}
	}, {
		safe: true,
		multi: true
	}, _);
	exports.tracer && exports.tracer("Update script F_120636_office_document_upload to version: 1 executed");
};

exports.dataUpdate = function(_, db, actualVersion, targetVersion) {
	// force log: always
	exports.tracer = console.log;
	//
	_scripts.slice(actualVersion + 1, targetVersion + 1).forEach_(_, function(_, sequence) {
		sequence && sequence(_, db);
	});
};

exports.metadata = {
	fileId: "c520bdf48bf4", // this id MUST never change and MUST be unique over all update scripts
	description: "F_120636_office_document_upload update script" // !important, some description, optional and can change
};
