// @sage/syracuse-lib/src/collaboration/updateScripts/updateScript_2017Rx.ts
"use strict";
exports.tracer; 
import * as config from 'config';
import * as adminHelpers from '../../collaboration/helpers';
var _scripts = [];

_scripts[1] = function(_, db) {
	exports.tracer && exports.tracer("Executing update script to version: 1; Set endpoint as initialized");
	var coll = db.db.collection("EndPoint", _);
	var elts = coll.find({}).toArray(_);

	if (elts && elts.length > 0) {
		elts.forEach_(_, function(_, e) {
			exports.tracer && exports.tracer("Update endpoint " + e.name);
			coll.update({
				_id: e._id
			}, {
				$set: {
					_updDate: new Date(),
					initialized: true
				}
			}, {
				safe: true,
				multi: true
			}, _);

		});
	}
	exports.tracer && exports.tracer("Update script to version: 1 executed");
};

export function dataUpdate(_, db, actualVersion, targetVersion) {
	// force log: always
	exports.tracer = console.error;
	//
	_scripts.slice(actualVersion + 1, targetVersion + 1).forEach_(_, function(_, sequence) {
		sequence && sequence(_, db);
	});
};

export const metadata = {
	fileId: "99861bfeba4e", // this id MUST never change and MUST be unique over all update scripts
	description: "2017Rx update script" // !important, some description, optional and can change
};
