// @sage/syracuse-lib/src/collaboration/updateScripts/updateScript_11-P02.ts
"use strict";

import { helpers } from '@sage/syracuse-core';

exports.tracer; // = console.log;
var _scripts = [];

_scripts[1] = function(_, db) {
	exports.tracer && exports.tracer("Executing update script to version: 1; add 'accessibility' css theme");
	var collTheme = db.db.collection("Theme", _);
	var themes = collTheme.find({
		code: "accessibility"
	}).toArray(_);

	var theme = {
		"code": "accessibility",
		"description": "accessibility",
		"cssFiles": [{
			"path": "accessibility",
			"_creUser": "admin",
			"_creDate": new Date(),
			"_uuid": helpers.uuid.generate("-"),
			"_updUser": "admin",
			"_updDate": new Date(),
			"_index": 0
		}],
		"_creUser": "admin",
		"_creDate": new Date(),
		"_updUser": "admin",
		"_updDate": new Date()
	};

	if (themes && themes.length) {
		collTheme.update({
			_id: themes[0]._id
		}, {
			$set: theme
		}, {
			safe: true,
			multi: true
		}, _);
	} else {
		theme._id = helpers.uuid.generate("-");
		collTheme.insert(theme, _);
	}
	exports.tracer && exports.tracer("Update script to version: 1 executed");
};

exports.dataUpdate = function(_, db, actualVersion, targetVersion) {
	// force log: always
	exports.tracer = console.error;
	//
	_scripts.slice(actualVersion + 1, targetVersion + 1).forEach_(_, function(_, sequence) {
		sequence && sequence(_, db);
	});
};

exports.metadata = {
	fileId: "a14fd1e5ee42", // this id MUST never change and MUST be unique over all update scripts
	description: "V11 P2 update script" // !important, some description, optional and can change
};
