// syracuse-main/lib/nodeconfig.js

var config = {};

try {
	config = require("../../../nodelocal").config || {};
} 
catch (ex) {
	console.log(ex);
}

config.collaboration = {
	contract: "syracuse",
	application: "collaboration",
	dataset: "syracuse"
}

config.sdata = {
	pubkeyName : "SYRASVR1",
	signAlgorithm: "RSA-SHA256",
	endpoints: [{
		contract: require("syracuse-collaboration/lib/contract").contract,
		datasets: {
			syracuse: {
				driver: "mongodb",
				port: 27017,
				hostname: config.sqlHostName || "localhost"
			}
		}
	},{
		contract: require("syracuse-search/lib/contract").contract,
		datasets: {
			syracuse: {
				driver: "mongodb",
				port: 27017,
				hostname: config.sqlHostName || "localhost"
			}
		}
	},{
		contract: require("sprint1/lib/settings/contract").contract,
		datasets: {
			sprint1: {
				driver: "mongodb",
				port: 27017,
				hostname: config.sqlHostName || "localhost"
//				login: "root",
//				password: config.sqlPassword || ""
			}
		}
	},{
		contract: require("zz/lib/test/contract").contract,
		datasets: {
			zz: {
				driver: "mongodb",
				port: 27017,
				hostname: config.sqlHostName || "localhost"
//				login: "root",
//				password: config.sqlPassword || ""
			}
		}
	}]
};

config.x3 = {
	solutions: {
		sodaix02: {
			"solution": {
		        "name": "SOLSUPV6",
		        "mainPort": 1801
		    },
		    "application": {
		        "server": "sodaix02.sagefr.adinternal.com",
		        "mainPort": 80
		    },
		    "runtimes": [
		        {
		            "name": "SERV_TRT1",
		            "server": "sodaix03",
		            "mainPort": 1906
		        },
		        {
		            "name": "SERV_TRT3",
		            "server": "sodaix04",
		            "mainPort": 1902
		        }
		    ],
		    "webServers": [
/*   		        {
		            "name": "ECLVARCIAT",
		            "server": "eclvarciat.sagefr.adinternal.com",
		            "mainPort": 28880,
		            "adminPort": 1818
		        }
				{
                   "name": "ecfdalbo02",
                   "server": "ecfdalbo02",
                   "mainPort": 18780,
                   "adminPort": 1818
               }*/
				{
                   "name": "ecvmdevrhx64",
                   "server": "ecvmdevrhx64",
                   "mainPort": 18780,
                   "adminPort": 1818
               }
		    ],
		    "javaServers": [
		        {
		            "name": "JAVA01",
		            "server": "jav01",
		            "mainPort": 1243
		        }
		    ],
		    "folders": [
		        {
		            "name": "X3",
		            "version": "6.2",
		            "release": "33",
		            "patch": "15",
		            "languages": [ "fr-FR","en-GB","zh","de","it","pt","es","en-US","pl","ru" ]
		        },
		        {
		            "name": "DEMO",
		            "version": "6.2",
		            "release": "33",
		            "patch": "15",
		            "languages": [ "fr-FR","en-US","es" ]
		        }
		    ]
		},
		"172.28.16.106": {
			"solution": {
		        "name": "SOLSUPV6",
		        "mainPort": 1801
		    },
		    "application": {
		        "server": "sodaix02.sagefr.adinternal.com",
		        "mainPort": 80
		    },
		    "runtimes": [
		        {
		            "name": "SERV_TRT1",
		            "server": "sodaix03",
		            "mainPort": 1906
		        },
		        {
		            "name": "SERV_TRT3",
		            "server": "sodaix04",
		            "mainPort": 1902
		        }
		    ],
		    "webServers": [
/*		        {
		            "name": "ECLVARCIAT",
		            "server": "eclvarciat.sagefr.adinternal.com",
		            "mainPort": 28880,
		            "adminPort": 1818
		        }
	
				{
				    "name": "ecfdalbo02",
				    "server": "ecfdalbo02",
				    "mainPort": 18780,
				    "adminPort": 1818
				}	        
				
*/
				{
                   "name": "ecvmdevrhx64",
                   "server": "ecvmdevrhx64",
                   "mainPort": 18780,
                   "adminPort": 1818
               }
		    ],
		    "javaServers": [
		        {
		            "name": "JAVA01",
		            "server": "jav01",
		            "mainPort": 1243
		        }
		    ],
		    "folders": [
		        {
		            "name": "X3",
		            "version": "6.2",
		            "release": "33",
		            "patch": "15",
		            "languages": [ "fr-FR","en-GB","zh","de","it","pt","es","en-US","pl","ru" ]
		        },
		        {
		            "name": "DEMO",
		            "version": "6.2",
		            "release": "33",
		            "patch": "15",
		            "languages": [ "fr-FR","en-US","es" ]
		        }
		    ]
		}
	}
}

exports.config = config;
