// @sage/syracuse-ui/lib/authoring/fieldAddins/choiceFieldAddin.js
"use strict";
exports.load = function(designer) {
	designer.$prototype.$properties.$choiceField$format = {
		$type: "application/x-choice",
		$title: syra_local.fchoice_formatTitle,
		$value: {
			$type: "application/x-string",
			$enum: [{
				$value: "$combo",
				$title: syra_local.fchoice_formatList
			}, {
				$value: "$radios",
				$title: syra_local.fchoice_formatRadios
			}]
		}
	};
	designer.$prototype.$properties.$choiceField$choiceLayout = {
		$type: "application/x-choice",
		$title: syra_local.fchoice_layout_title,
		$value: {
			$type: "application/x-string",

			$enum: [{
				$value: "row",
				$title: syra_local.fchoice_layout_1_row
			}, {
				$value: "1",
				$title: syra_local.fchoice_layout_1_col
			}, {
				$value: "2",
				$title: syra_local.fchoice_layout_2_col
			}, {
				$value: "3",
				$title: syra_local.fchoice_layout_3_col
			}, {
				$value: "4",
				$title: syra_local.fchoice_layout_4_col
			}]
		}
	};
};

exports.onShow = function(designer) {
	var item = designer.designedItem;
	if (item.$authoringType == "choiceField") {
		designer.applyChange({
			$choiceField$format: item.$item.$format,
			$choiceField$choiceLayout: item.$item.$choiceLayout
		});
	}
};

exports.onFormUpdateDelta = function(designer, metaData) {
	if (designer.designedItem.$authoringType == "choiceField") {
		if (metaData.$choiceField$format != undefined || metaData.$choiceField$choiceLayout != undefined) {
			designer.designedItem.applyDesignMeta({
				$format: metaData.$choiceField$format,
				$choiceLayout: metaData.$choiceField$choiceLayout
			}, true);
		}
	}
};

exports.addToSection = function(designer, section) {
	section.properties.push(designer.addItem(section.body, {
		$authoringType: "choiceField",
		$authoringEditMode: true,
		$bind: "$choiceField$format",
		$isEditMode: true,
		$choiceLayout: "1"
	}, designer));
	section.properties.push(designer.addItem(section.body, {
		$authoringType: "choiceField",
		$authoringEditMode: true,
		$bind: "$choiceField$choiceLayout",
		$format: "$combo",
		$isEditMode: true
	}, designer));
};
