// @sage/syracuse-ui/lib/authoring/fieldAddins/htmlFielddAddin.js
"use strict";
exports.load = function(designer) {
	designer.$prototype.$properties.htmhField$title = {
		$type: "application/x-string",
		$title: syra_local.htmlVignetteTitle,
		$displayLength: 200
	};
	designer.$prototype.$properties.htmhField$url = {
		$type: "application/x-string",
		$isHidden: true,
		$title: syra_local.htmlVignetteUrl,
		$displayLength: 200
	};
};

exports.onShow = function(designer) {
	var item = designer.designedItem;
	if (item.$authoringType == "htmlVignetteField") {
		designer.applyChange({
			htmhField$title: item.getTitle() || item.$item.$title,
			htmhField$url: item.$contentUrl
		});
	}
};


exports.onFormUpdateDelta = function(designer, metaData) {
	if (designer.designedItem.$authoringType == "htmlVignetteField") {
		if (metaData.htmhField$title != undefined || metaData.htmhField$url != undefined) {
			designer.designedItem.applyDesignMeta({
				$title: metaData.htmhField$title,
				$url: metaData.htmhField$url
			}, true);
		}
	}
};

exports.addToSection = function(designer, section) {
	section.properties.push(designer.addItem(section.body, {
		$authoringType: "htmlVignetteField",
		$bind: "htmhField$title",
		$isEditMode: true
	}, designer));
	section.properties.push(designer.addItem(section.body, {
		$authoringType: "htmlVignetteField",
		$bind: "htmhField$url",
		$isEditMode: true
	}, designer));
};
