// syracuse-ui/lib/field/childObject/childObjectFieldContextMenu.js
"use strict";
var helpers = require('syracuse-core/lib/helpers');
var ContextMenu = require('syracuse-ui/lib/field/menus/contextMenu').ContextMenu;


function ChildObjectFieldContextMenu(){
}

exports.ChildObjectFieldContextMenu = helpers.defineClass(ChildObjectFieldContextMenu, ContextMenu, {
    clearMenuItems: function(){
        var self = this;
        self._isMenuBtnVisible = false;
        self._menusBtnSlot.style.display = "none";
    },
    create: function(field){
        var self = this
        if (field.list) {
            self.$skinMenus = "s-list-record-menus";
            self.$skinActionSlot = field.list.$skin + "-record-actions";
        }
        else {
            self.$skinMenus = "s-list-menus";
            self.$skinActionSlot = "s-list-actions";
        }
        ContextMenu.prototype.create.call(this, field, {
            onBeforeAddMenuItem: function($menuItem){
                if (!self._isMenuBtnVisible) {
                    self._menusBtnSlot.style.display = "";
                    self._isMenuBtnVisible = true;
                    if ($menuItem.$bind == "$lazyload") {
                        var title;
                        if (self.field.list) {
                            title = self.field.list._localize.mn_actions;
                        }
                        else {
                            title = self.field._localize.mn_actions;
                            $(self._menusBtn).text(title);
                        }
                        self._menusBtn.setAttribute("title", title);
                        return false;
                    }
                }
            }
        });
    },
    appendOpener: function(){
        this._actionsSlot = document.createElement("div");
        this._actionsSlot.className = this.$skinActionSlot;
        this.field._contextMenusSlot.appendChild(this._actionsSlot);
        this._menusBtn = document.createElement("a");
        var title;
        if (this.field.list) {
            this._menusBtn.className = "s-list-record-menus-btn s-close";
            title = this.field.list._localize.mn_more;
        }
        else {
            this._menusBtn.className = "s-list-menus-btn s-close";
            title = this.field._localize.mn_more;
            $(this._menusBtn).text(title);
        }
        this._menusBtn.setAttribute("title", title);
        
        this._isMenuBtnVisible = false;
        this._menusBtn.setAttribute("id", this.field.id + "-menu-btn");
        (this._menusBtnSlot = document.createElement("div")).className = "s-list-quick-actions-slot";
        this._menusBtnSlot.style.display = "none";
        return this._actionsSlot.appendChild(this._menusBtnSlot).appendChild(this._menusBtn);
    },
    removeOpener: function(){
        if (this._menusBtn) {
            document.site.emptyDom(this.field._contextMenusSlot);
            delete this._menusBtn;
        }
    }
});
